// @bitgo/sdk-coin-islm/dist/test/unit/islm.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sinon_1 = __importDefault(require("sinon"));
const src_1 = require("../../src");
const utils_1 = __importDefault(require("../../src/lib/utils"));
const islm_1 = require("../resources/islm");
const should = require("should");
describe('Islm', function () {
    let bitgo;
    let basecoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('islm', src_1.Islm.createInstance);
        bitgo.safeRegister('tislm', src_1.Tislm.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('tislm');
    });
    it('should return the right info', function () {
        const islm = bitgo.coin('islm');
        const tislm = bitgo.coin('tislm');
        islm.getChain().should.equal('islm');
        islm.getFamily().should.equal('islm');
        islm.getFullName().should.equal('Islamic Coin');
        islm.getBaseFactor().should.equal(1e18);
        tislm.getChain().should.equal('tislm');
        tislm.getFamily().should.equal('islm');
        tislm.getFullName().should.equal('Testnet Islamic Coin');
        tislm.getBaseFactor().should.equal(1e18);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const addressDetails = basecoin.getAddressDetails(islm_1.address.noMemoIdAddress);
            addressDetails.address.should.equal(islm_1.address.noMemoIdAddress);
            should.not.exist(addressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const addressDetails = basecoin.getAddressDetails(islm_1.address.validMemoIdAddress);
            addressDetails.address.should.equal(islm_1.address.validMemoIdAddress.split('?')[0]);
            addressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                basecoin.getAddressDetails(islm_1.address.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const receiveAddress = {
                address: 'haqq1g3g6nfmqf3f9lmdhf5g84pu7ustw8jt7tvrfzc?memoId=7',
                coinSpecific: {
                    rootAddress: 'haqq1g3g6nfmqf3f9lmdhf5g84pu7ustw8jt7tvrfzc',
                    memoID: '7',
                },
            };
            const isValid = await basecoin.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(utils_1.default.isValidAddress(islm_1.address.address1), true);
            should.equal(utils_1.default.isValidAddress(islm_1.address.address2), true);
            should.equal(utils_1.default.isValidAddress(islm_1.address.address3), false);
            should.equal(utils_1.default.isValidAddress(islm_1.address.address4), true);
            should.equal(utils_1.default.isValidAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidAddress(undefined), false);
            should.equal(utils_1.default.isValidAddress(''), false);
            should.equal(utils_1.default.isValidAddress(islm_1.address.validMemoIdAddress), true);
            should.equal(utils_1.default.isValidAddress(islm_1.address.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(utils_1.default.isValidValidatorAddress(islm_1.address.validatorAddress1), true);
            should.equal(utils_1.default.isValidValidatorAddress(islm_1.address.validatorAddress2), true);
            should.equal(utils_1.default.isValidValidatorAddress(islm_1.address.validatorAddress3), false);
            should.equal(utils_1.default.isValidValidatorAddress(islm_1.address.validatorAddress4), false);
            should.equal(utils_1.default.isValidValidatorAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidValidatorAddress(undefined), false);
            should.equal(utils_1.default.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: islm_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: islm_1.TEST_SEND_TX.recipient,
                        amount: islm_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify delegate transaction', async function () {
            const txPrebuild = {
                txHex: islm_1.TEST_DELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: islm_1.TEST_DELEGATE_TX.validator,
                        amount: islm_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify undelegate transaction', async function () {
            const txPrebuild = {
                txHex: islm_1.TEST_UNDELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: islm_1.TEST_UNDELEGATE_TX.validator,
                        amount: islm_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify withdraw rewards transaction', async function () {
            const txPrebuild = {
                txHex: islm_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: islm_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: islm_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: islm_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: islm_1.TEST_SEND_TX.recipient,
                        amount: islm_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: islm_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: islm_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a delegate transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: islm_1.TEST_DELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: islm_1.TEST_DELEGATE_TX.hash,
                outputs: [
                    {
                        address: islm_1.TEST_DELEGATE_TX.validator,
                        amount: islm_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: islm_1.TEST_DELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: islm_1.TEST_DELEGATE_TX.gasBudget.amount[0].amount },
                type: 13,
            });
        });
        it('should explain a undelegate transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: islm_1.TEST_UNDELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: islm_1.TEST_UNDELEGATE_TX.hash,
                outputs: [
                    {
                        address: islm_1.TEST_UNDELEGATE_TX.validator,
                        amount: islm_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: islm_1.TEST_UNDELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: islm_1.TEST_UNDELEGATE_TX.gasBudget.amount[0].amount },
                type: 17,
            });
        });
        it('should explain a withdraw transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: islm_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: islm_1.TEST_WITHDRAW_REWARDS_TX.hash,
                outputs: [
                    {
                        address: islm_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
                outputAmount: undefined,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: islm_1.TEST_WITHDRAW_REWARDS_TX.gasBudget.amount[0].amount },
                type: 15,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: islm_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: islm_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: islm_1.TEST_TX_WITH_MEMO.recipient,
                        amount: islm_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: islm_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: islm_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: islm_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await basecoin.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await basecoin.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        it('should parse a transfer transaction', async function () {
            const transferInputsResponse = {
                address: islm_1.TEST_SEND_TX.recipient,
                amount: new bignumber_js_1.default(islm_1.TEST_SEND_TX.sendAmount).plus(islm_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
            };
            const transferOutputsResponse = {
                address: islm_1.TEST_SEND_TX.recipient,
                amount: islm_1.TEST_SEND_TX.sendAmount,
            };
            const parsedTransaction = await basecoin.parseTransaction({ txHex: islm_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Islm.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await basecoin
                .parseTransaction({ txHex: islm_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
});
