// @bitgo/wasm-utxo/dist/cjs/js/testutils/descriptor/descriptors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUnspendableKey = getUnspendableKey;
exports.getDefaultXPubs = getDefaultXPubs;
exports.getPsbtParams = getPsbtParams;
exports.getDescriptorNode = getDescriptorNode;
exports.containsKey = containsKey;
exports.getTapLeafScripts = getTapLeafScripts;
exports.getDescriptor = getDescriptor;
exports.getDescriptorMap = getDescriptorMap;
/**
 * Descriptor test utilities for building common descriptor templates.
 * Ported from @bitgo/utxo-core/testutil/descriptor/descriptors.ts.
 */
const assert_1 = __importDefault(require("assert"));
const bip32_js_1 = require("../../bip32.js");
const index_js_1 = require("../../index.js");
const keys_js_1 = require("../keys.js");
/**
 * Get the BIP-341 "Nothing Up My Sleeve" (NUMS) unspendable key.
 * This is the x-only public key with unknown discrete logarithm
 * constructed by hashing the uncompressed secp256k1 base point G.
 *
 * @see https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs
 */
function getUnspendableKey() {
    return "50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0";
}
function getDefaultXPubs(seed) {
    return (0, keys_js_1.getKeyTriple)(seed ?? "default").map((k) => k.neutered().toBase58());
}
function toDescriptorMap(v) {
    return new Map(Object.entries(v).map(([k, v]) => [k, index_js_1.Descriptor.fromString(v, "derivable")]));
}
function toXPub(k, path) {
    if (typeof k === "string") {
        return k + "/" + path;
    }
    return k.neutered().toBase58() + "/" + path;
}
function toPlain(k, { xonly = false } = {}) {
    if (typeof k === "string") {
        if (k.startsWith("xpub") || k.startsWith("xprv")) {
            return toPlain(bip32_js_1.BIP32.fromBase58(k), { xonly });
        }
        return k;
    }
    return toHex(k.publicKey.subarray(xonly ? 1 : 0));
}
function toHex(bytes) {
    return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}
function toXOnly(k) {
    return toPlain(k, { xonly: true });
}
function multiArgs(m, n, keys, path) {
    if (n < m) {
        throw new Error(`Cannot create ${m} of ${n} multisig`);
    }
    if (keys.length < n) {
        throw new Error(`Not enough keys for ${m} of ${n} multisig: keys.length=${keys.length}`);
    }
    keys = keys.slice(0, n);
    return [m, ...keys.map((k) => toXPub(k, path))];
}
function getPsbtParams(t) {
    switch (t) {
        case "Wsh2Of3CltvDrop":
            return { locktime: 1 };
        default:
            return {};
    }
}
function getDescriptorNode(template, keys = getDefaultXPubs(), path = "0/*") {
    switch (template) {
        case "Wsh2Of3":
            return {
                wsh: { multi: multiArgs(2, 3, keys, path) },
            };
        case "Wsh2Of3CltvDrop": {
            const { locktime } = getPsbtParams(template);
            (0, assert_1.default)(locktime);
            return {
                wsh: {
                    and_v: [{ "r:after": locktime }, { multi: multiArgs(2, 3, keys, path) }],
                },
            };
        }
        case "Wsh2Of2":
            return {
                wsh: { multi: multiArgs(2, 2, keys, path) },
            };
        case "Tr2Of3-NoKeyPath":
            return {
                tr: [getUnspendableKey(), { multi_a: multiArgs(2, 3, keys, path) }],
            };
        case "Tr1Of3-NoKeyPath-Tree":
            return {
                tr: [
                    getUnspendableKey(),
                    [
                        { pk: toXPub(keys[0], path) },
                        [{ pk: toXPub(keys[1], path) }, { pk: toXPub(keys[2], path) }],
                    ],
                ],
            };
        case "Tr1Of3-NoKeyPath-Tree-Plain":
            return {
                tr: [
                    getUnspendableKey(),
                    [{ pk: toXOnly(keys[0]) }, [{ pk: toXOnly(keys[1]) }, { pk: toXOnly(keys[2]) }]],
                ],
            };
    }
    throw new Error(`Unknown descriptor template: ${template}`);
}
function getTapLeafScriptNodes(t) {
    if (Array.isArray(t)) {
        if (t.length !== 2) {
            throw new Error(`expected tuple, got: ${JSON.stringify(t)}`);
        }
        return t.map((v) => (Array.isArray(v) ? getTapLeafScriptNodes(v) : v)).flat();
    }
    if (typeof t === "object") {
        const node = t;
        if (!("tr" in node)) {
            throw new Error(`TapLeafScripts are only supported for Taproot descriptors, got: ${JSON.stringify(t)}`);
        }
        if (!Array.isArray(node.tr) || node.tr.length !== 2) {
            throw new Error(`expected tuple, got: ${JSON.stringify(node.tr)}`);
        }
        const tapscript = node.tr[1];
        if (!Array.isArray(tapscript)) {
            throw new Error(`expected tapscript to be an array, got: ${JSON.stringify(tapscript)}`);
        }
        return getTapLeafScriptNodes(tapscript);
    }
    throw new Error(`Invalid input: ${JSON.stringify(t)}`);
}
function containsKey(script, key) {
    if (script instanceof index_js_1.Miniscript) {
        script = index_js_1.ast.fromMiniscript(script);
    }
    if ("pk" in script) {
        return script.pk === toXOnly(key);
    }
    throw new Error(`Unsupported script type: ${JSON.stringify(script)}`);
}
function getTapLeafScripts(d) {
    return getTapLeafScriptNodes(index_js_1.ast.fromDescriptor(d)).map((n) => index_js_1.Miniscript.fromString(index_js_1.ast.formatNode(n), "tap").toString());
}
function getDescriptor(template, keys = getDefaultXPubs(), path = "0/*") {
    return index_js_1.Descriptor.fromStringDetectType(index_js_1.ast.formatNode(getDescriptorNode(template, keys, path)));
}
function getDescriptorMap(template, keys = getDefaultXPubs()) {
    return toDescriptorMap({
        external: getDescriptor(template, keys, "0/*").toString(),
        internal: getDescriptor(template, keys, "1/*").toString(),
    });
}
