// @bitgo/wasm-utxo/dist/cjs/js/testutils/descriptor/mockPsbt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockDerivedDescriptorWalletOutput = mockDerivedDescriptorWalletOutput;
exports.mockPsbt = mockPsbt;
exports.mockPsbtDefault = mockPsbtDefault;
exports.mockPsbtDefaultWithDescriptorTemplate = mockPsbtDefaultWithDescriptorTemplate;
const index_js_1 = require("../../descriptorWallet/index.js");
const descriptors_js_1 = require("./descriptors.js");
function mockOutputId(id) {
    const hash = id?.hash ?? "0101010101010101010101010101010101010101010101010101010101010101";
    const vout = id?.vout ?? 0;
    return { hash, vout };
}
function mockDerivedDescriptorWalletOutput(descriptor, outputParams = {}) {
    const { value = BigInt(1e6) } = outputParams;
    const { hash, vout } = mockOutputId(outputParams.id);
    return {
        hash,
        index: vout,
        witnessUtxo: {
            script: (0, index_js_1.createScriptPubKeyFromDescriptor)(descriptor, undefined),
            value,
        },
        descriptor,
        selectTapLeafScript: outputParams.selectTapLeafScript,
        sequence: outputParams.sequence,
    };
}
function tryDeriveAtIndex(descriptor, index) {
    return descriptor.hasWildcard() ? descriptor.atDerivationIndex(index) : descriptor;
}
function mockPsbt(inputs, outputs, params = {}) {
    return (0, index_js_1.createPsbt)(params, inputs.map((i) => mockDerivedDescriptorWalletOutput(tryDeriveAtIndex(i.descriptor, i.index), i)), outputs.map((o) => {
        const derivedDescriptor = tryDeriveAtIndex(o.descriptor, o.index);
        return {
            script: (0, index_js_1.createScriptPubKeyFromDescriptor)(derivedDescriptor, undefined),
            value: o.value,
            descriptor: o.external ? undefined : derivedDescriptor,
        };
    }));
}
function mockPsbtDefault({ descriptorSelf = (0, descriptors_js_1.getDescriptor)("Wsh2Of3", (0, descriptors_js_1.getDefaultXPubs)("a")), descriptorOther = (0, descriptors_js_1.getDescriptor)("Wsh2Of3", (0, descriptors_js_1.getDefaultXPubs)("b")), params = {}, } = {}) {
    return mockPsbt([
        { descriptor: descriptorSelf, index: 0 },
        { descriptor: descriptorSelf, index: 1, id: { vout: 1 } },
    ], [
        {
            descriptor: descriptorOther,
            index: 0,
            value: BigInt(4e5),
            external: true,
        },
        { descriptor: descriptorSelf, index: 0, value: BigInt(4e5) },
    ], params);
}
function mockPsbtDefaultWithDescriptorTemplate(t, params = {}) {
    return mockPsbtDefault({
        descriptorSelf: (0, descriptors_js_1.getDescriptor)(t, (0, descriptors_js_1.getDefaultXPubs)("a")),
        descriptorOther: (0, descriptors_js_1.getDescriptor)(t, (0, descriptors_js_1.getDefaultXPubs)("b")),
        params: { ...(0, descriptors_js_1.getPsbtParams)(t), ...params },
    });
}
