// @bitgo/sdk-coin-icp/dist/src/icp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Icp = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const principal_1 = require("@dfinity/principal");
const axios_1 = __importDefault(require("axios"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const crypto_1 = require("crypto");
const mpc = __importStar(require("@bitgo/sdk-lib-mpc"));
const iface_1 = require("./lib/iface");
const transactionBuilderFactory_1 = require("./lib/transactionBuilderFactory");
const utils_1 = __importDefault(require("./lib/utils"));
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const icpAgent_1 = require("./lib/icpAgent");
/**
 * Class representing the Internet Computer (ICP) coin.
 * Extends the BaseCoin class and provides specific implementations for ICP.
 *
 * @see {@link https://internetcomputer.org/}
 * @see {@link https://internetcomputer.org/docs/current/developer-docs/defi/rosetta/icp_rosetta/data_api/}
 */
class Icp extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Icp(bitgo, staticsCoin);
    }
    getChain() {
        return 'icp';
    }
    getBaseChain() {
        return 'icp';
    }
    getFamily() {
        return this._staticsCoin.family;
    }
    getFullName() {
        return 'Internet Computer';
    }
    getBaseFactor() {
        return Math.pow(10, this._staticsCoin.decimalPlaces);
    }
    async explainTransaction(params) {
        const factory = this.getBuilderFactory();
        const txBuilder = await factory.from(params.transactionHex);
        const transaction = await txBuilder.build();
        if (params.signableHex !== undefined) {
            const generatedSignableHex = txBuilder.transaction.payloadsData.payloads[0].hex_bytes;
            if (generatedSignableHex !== params.signableHex) {
                throw new Error('generated signableHex is not equal to params.signableHex');
            }
        }
        return transaction.explainTransaction();
    }
    async verifyTransaction(params) {
        const { txParams, txPrebuild } = params;
        const txHex = txPrebuild?.txHex;
        if (!txHex) {
            throw new Error('txHex is required');
        }
        const txHexParams = {
            transactionHex: txHex,
        };
        if (txPrebuild.txInfo && txPrebuild.txInfo !== undefined && typeof txPrebuild.txInfo === 'string') {
            txHexParams.signableHex = txPrebuild.txInfo;
        }
        const explainedTx = await this.explainTransaction(txHexParams);
        if (Array.isArray(txParams.recipients) && txParams.recipients.length > 0) {
            if (txParams.recipients.length > 1) {
                throw new Error(`${this.getChain()} doesn't support sending to more than 1 destination address within a single transaction. Try again, using only a single recipient.`);
            }
            (0, assert_1.default)(explainedTx.outputs.length === 1, 'Tx outputs does not match with expected txParams recipients');
            const output = explainedTx.outputs[0];
            const recipient = txParams.recipients[0];
            (0, assert_1.default)(typeof recipient.address === 'string' &&
                typeof output.address === 'string' &&
                output.address === recipient.address &&
                (0, bignumber_js_1.default)(output.amount).eq((0, bignumber_js_1.default)(recipient.amount)), 'Tx outputs does not match with expected txParams recipients');
        }
        return true;
    }
    /**
     * Verify that an address belongs to this wallet.
     *
     * @param {TssVerifyIcpAddressOptions} params - Verification parameters
     * @returns {Promise<boolean>} True if address belongs to wallet
     * @throws {InvalidAddressError} If address format is invalid or doesn't match derived address
     * @throws {Error} If invalid wallet version or missing parameters
     */
    async isWalletAddress(params) {
        const { address, rootAddress, walletVersion } = params;
        if (!this.isValidAddress(address)) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        let addressToVerify = address;
        if (walletVersion === 1) {
            if (!rootAddress) {
                throw new Error('rootAddress is required for wallet version 1');
            }
            const extractedRootAddress = utils_1.default.validateMemoAndReturnRootAddress(address);
            if (!extractedRootAddress || extractedRootAddress === address) {
                throw new Error('memoId is required for wallet version 1 addresses');
            }
            if (extractedRootAddress.toLowerCase() !== rootAddress.toLowerCase()) {
                throw new sdk_core_1.UnexpectedAddressError(`address validation failure: expected ${rootAddress} but got ${extractedRootAddress}`);
            }
            addressToVerify = rootAddress;
        }
        const indexToVerify = walletVersion === 1 ? 0 : params.index;
        const result = await (0, sdk_core_1.verifyMPCWalletAddress)({ ...params, address: addressToVerify, index: indexToVerify, keyCurve: 'secp256k1' }, this.isValidAddress.bind(this), (pubKey) => utils_1.default.getAddressFromPublicKey(pubKey));
        if (!result) {
            throw new sdk_core_1.UnexpectedAddressError(`address validation failure: address ${addressToVerify} is not a wallet address`);
        }
        return true;
    }
    async parseTransaction(params) {
        return {};
    }
    /**
     * Generate a new keypair for this coin.
     * @param seed Seed from which the new keypair should be generated, otherwise a random seed is used
     */
    generateKeyPair(seed) {
        return utils_1.default.generateKeyPair(seed);
    }
    isValidAddress(address) {
        return utils_1.default.isValidAddress(address);
    }
    async signTransaction(params) {
        const txHex = params?.txPrebuild?.txHex;
        const privateKey = params?.prv;
        if (!txHex) {
            throw new sdk_core_1.SigningError('missing required txPrebuild parameter: params.txPrebuild.txHex');
        }
        if (!privateKey) {
            throw new sdk_core_1.SigningError('missing required prv parameter: params.prv');
        }
        const factory = this.getBuilderFactory();
        const txBuilder = await factory.from(params.txPrebuild.txHex);
        txBuilder.sign({ key: params.prv });
        txBuilder.combine();
        const serializedTx = txBuilder.transaction.toBroadcastFormat();
        return {
            txHex: serializedTx,
        };
    }
    isValidPub(key) {
        return utils_1.default.isValidPublicKey(key);
    }
    isValidPrv(key) {
        return utils_1.default.isValidPrivateKey(key);
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /** @inheritDoc **/
    getHashFunction() {
        return (0, crypto_1.createHash)('sha256');
    }
    getAddressFromPublicKey(hexEncodedPublicKey) {
        return utils_1.default.getAddressFromPublicKey(hexEncodedPublicKey);
    }
    /** @inheritDoc **/
    getPublicNodeUrl() {
        return sdk_core_1.Environments[this.bitgo.getEnv()].icpNodeUrl;
    }
    /** @inheritDoc **/
    // this method calls the public node to broadcast the transaction and not the rosetta node
    async broadcastTransaction(payload) {
        const endpoint = this.getPublicNodeBroadcastEndpoint();
        try {
            const bodyBytes = utils_1.default.blobFromHex(payload.serializedSignedTransaction);
            const response = await axios_1.default.post(endpoint, bodyBytes, {
                headers: { 'Content-Type': 'application/cbor' },
                responseType: 'arraybuffer', // This ensures you get a Buffer, not a string
            });
            if (response.status !== 200) {
                throw new Error(`Transaction broadcast failed with status: ${response.status} - ${response.statusText}`);
            }
            const decodedResponse = utils_1.default.cborDecode(response.data);
            if (decodedResponse.status === 'replied') {
                // it is considered a success because ICP returns response in a CBOR map with a status of 'replied'
                return {}; // returned empty object as ICP does not return a txid
            }
            else {
                throw new Error(`Unexpected response status from node: ${decodedResponse.status}`);
            }
        }
        catch (error) {
            throw new Error(`Transaction broadcast error: ${error?.message || JSON.stringify(error)}`);
        }
    }
    getPublicNodeBroadcastEndpoint() {
        const nodeUrl = this.getPublicNodeUrl();
        const ledgerCanisterId = this._staticsCoin.network.type === statics_1.NetworkType.TESTNET ? iface_1.TESTNET_LEDGER_CANISTER_ID : iface_1.LEDGER_CANISTER_ID;
        const principal = principal_1.Principal.fromUint8Array(ledgerCanisterId);
        const canisterIdHex = principal.toText();
        const endpoint = `${nodeUrl}${iface_1.PUBLIC_NODE_REQUEST_ENDPOINT}${canisterIdHex}/call`;
        return endpoint;
    }
    /**
     * Fetches the account balance for a given public key.
     * @param publicKeyHex - Hex-encoded public key of the account.
     * @returns Promise resolving to the account balance as a string.
     * @throws Error if the balance could not be fetched.
     */
    async getAccountBalance(publicKeyHex) {
        const principalId = utils_1.default.getPrincipalIdFromPublicKey(publicKeyHex).toText();
        const agent = new icpAgent_1.IcpAgent(this.getPublicNodeUrl(), this._staticsCoin);
        return agent.getBalance(principalId);
    }
    /**
     * Retrieves the current transaction fee data from the ICP public node.
     *
     * This method creates an instance of `IcpAgent` using the public node URL,
     * then queries the node for the current fee information.
     *
     * @returns A promise that resolves to a `BigNumber` representing the current transaction fee.
     * @throws Will propagate any errors encountered while communicating with the ICP node.
     */
    async getFeeData() {
        const agent = new icpAgent_1.IcpAgent(this.getPublicNodeUrl(), this._staticsCoin);
        return await agent.getFee();
    }
    getBuilderFactory() {
        return new transactionBuilderFactory_1.TransactionBuilderFactory(statics_1.coins.get(this.getBaseChain()));
    }
    /**
     * Generates an array of signatures for the provided payloads using MPC
     *
     * @param payloadsData - The data containing the payloads to be signed.
     * @param senderPublicKey - The public key of the sender in hexadecimal format.
     * @param userKeyShare - The user's key share as a Buffer.
     * @param backupKeyShare - The backup key share as a Buffer.
     * @param commonKeyChain - The common key chain identifier used for MPC signing.
     * @returns A promise that resolves to an array of `Signatures` objects, each containing the signing payload,
     *          signature type, public key, and the generated signature in hexadecimal format.
     */
    async signatures(payloadsData, senderPublicKey, userKeyShare, backupKeyShare, commonKeyChain) {
        try {
            const payload = payloadsData.payloads[0];
            const message = Buffer.from(payload.hex_bytes, 'hex');
            const messageHash = (0, crypto_1.createHash)('sha256').update(message).digest();
            const signature = await sdk_core_1.ECDSAUtils.signRecoveryMpcV2(messageHash, userKeyShare, backupKeyShare, commonKeyChain);
            const signaturePayload = {
                signing_payload: payload,
                signature_type: payload.signature_type,
                public_key: {
                    hex_bytes: senderPublicKey,
                    curve_type: iface_1.CurveType.SECP256K1,
                },
                hex_bytes: signature.r + signature.s,
            };
            return [signaturePayload];
        }
        catch (error) {
            throw new Error(`Error generating signatures: ${error.message || error}`);
        }
    }
    /**
     * Builds a funds recovery transaction without BitGo
     * @param params
     */
    async recover(params) {
        try {
            if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
                throw new Error('invalid recoveryDestination');
            }
            const isUnsignedSweep = !params.userKey && !params.backupKey && !params.walletPassphrase;
            let publicKey;
            let userKeyShare, backupKeyShare, commonKeyChain;
            const MPC = new sdk_core_1.Ecdsa();
            if (!isUnsignedSweep) {
                if (!params.userKey) {
                    throw new Error('missing userKey');
                }
                if (!params.backupKey) {
                    throw new Error('missing backupKey');
                }
                if (!params.walletPassphrase) {
                    throw new Error('missing wallet passphrase');
                }
                const userKey = params.userKey.replace(/\s/g, '');
                const backupKey = params.backupKey.replace(/\s/g, '');
                ({ userKeyShare, backupKeyShare, commonKeyChain } = await sdk_core_1.ECDSAUtils.getMpcV2RecoveryKeyShares(userKey, backupKey, params.walletPassphrase));
                publicKey = MPC.deriveUnhardened(commonKeyChain, iface_1.ROOT_PATH).slice(0, 66);
            }
            else {
                const bitgoKey = params.bitgoKey;
                if (!bitgoKey) {
                    throw new Error('missing bitgoKey');
                }
                const hdTree = new mpc.Secp256k1Bip32HdTree();
                const derivationPath = 'm/0';
                const derivedPub = hdTree.publicDerive({
                    pk: mpc.bigIntFromBufferBE(Buffer.from(bitgoKey.slice(0, 66), 'hex')),
                    chaincode: mpc.bigIntFromBufferBE(Buffer.from(bitgoKey.slice(66), 'hex')),
                }, derivationPath);
                publicKey = mpc.bigIntToBufferBE(derivedPub.pk).toString('hex');
            }
            if (!publicKey) {
                throw new Error('failed to derive public key');
            }
            const senderAddress = this.getAddressFromPublicKey(publicKey);
            const balance = await this.getAccountBalance(publicKey);
            const feeData = await this.getFeeData();
            const actualBalance = balance.minus(feeData);
            if (actualBalance.isLessThanOrEqualTo(0)) {
                throw new Error('Did not have enough funds to recover');
            }
            const factory = this.getBuilderFactory();
            const txBuilder = factory.getTransferBuilder();
            txBuilder.sender(senderAddress, publicKey);
            txBuilder.receiverId(params.recoveryDestination);
            txBuilder.amount(actualBalance.toString());
            if (params.memo !== undefined && utils_1.default.validateMemo(params.memo)) {
                txBuilder.memo(Number(params.memo));
            }
            await txBuilder.build();
            if (txBuilder.transaction.payloadsData.payloads.length === 0) {
                throw new Error('Missing payloads to generate signatures');
            }
            if (isUnsignedSweep) {
                return {
                    txHex: txBuilder.transaction.unsignedTransaction,
                    coin: this.getChain(),
                };
            }
            const signatures = await this.signatures(txBuilder.transaction.payloadsData, publicKey, userKeyShare, backupKeyShare, commonKeyChain);
            if (!signatures || signatures.length === 0) {
                throw new Error('Failed to generate signatures');
            }
            txBuilder.transaction.addSignature(signatures);
            txBuilder.combine();
            const broadcastableTxn = txBuilder.transaction.toBroadcastFormat();
            await this.broadcastTransaction({ serializedSignedTransaction: broadcastableTxn });
            const txId = txBuilder.transaction.id;
            const recoveredTransaction = {
                id: txId,
                tx: broadcastableTxn,
            };
            return recoveredTransaction;
        }
        catch (error) {
            throw new Error(`Error during ICP recovery: ${error.message || error}`);
        }
    }
    /** @inheritDoc */
    auditDecryptedKey({ multiSigType, prv, publicKey }) {
        if (multiSigType !== 'tss') {
            throw new Error('Unsupported multisigtype ');
        }
        (0, sdk_lib_mpc_1.auditEcdsaPrivateKey)(prv, publicKey);
    }
}
exports.Icp = Icp;
