// @bitgo/sdk-coin-vet/dist/src/lib/transaction/addressInitializationTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressInitializationTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const ethUtil = __importStar(require("ethereumjs-util"));
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const sdk_core_2 = require("@vechain/sdk-core");
const transaction_1 = require("./transaction");
class AddressInitializationTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.AddressInitialization;
    }
    get baseAddress() {
        return this._baseAddress;
    }
    set baseAddress(address) {
        this._baseAddress = address;
    }
    get feeAddress() {
        return this._feeAddress;
    }
    set feeAddress(address) {
        this._feeAddress = address;
    }
    get salt() {
        return this._salt;
    }
    set salt(salt) {
        this._salt = salt;
    }
    get initCode() {
        return this._initCode;
    }
    set initCode(initCode) {
        this._initCode = initCode;
    }
    get deployedAddress() {
        return this._deployedAddress;
    }
    set deployedAddress(address) {
        this._deployedAddress = address;
    }
    /** @inheritdoc */
    async build() {
        super.build();
        if (this._salt && this._initCode) {
            const saltBuffer = ethUtil.setLengthLeft(ethUtil.toBuffer(this._salt), 32);
            const { createForwarderParams, createForwarderTypes } = (0, abstract_eth_1.getCreateForwarderParamsAndTypes)(this._baseAddress, saltBuffer, this._feeAddress);
            // Hash the wallet base address and fee address if present with the given salt, so the address directly relies on the base address and fee address
            const calculationSalt = ethUtil.bufferToHex(ethereumjs_abi_1.default.soliditySHA3(createForwarderTypes, createForwarderParams));
            this.deployedAddress = (0, abstract_eth_1.calculateForwarderV1Address)(this._contract, calculationSalt, this._initCode);
        }
    }
    /** @inheritdoc */
    buildClauses() {
        this._clauses = [
            {
                to: this._contract,
                value: '0x0',
                data: this._transactionData,
            },
        ];
    }
    /** @inheritdoc */
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            data: this.transactionData,
            value: '0',
            sender: this.sender,
            to: this.contract,
            deployedAddress: this.deployedAddress,
        };
        return json;
    }
    /** @inheritdoc */
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set data from clauses
            this.contract = body.clauses[0]?.to || '0x0';
            this.transactionData = body.clauses[0]?.data || '0x0';
            this.type = sdk_core_1.TransactionType.AddressInitialization;
            const { baseAddress, addressCreationSalt, feeAddress } = (0, abstract_eth_1.decodeForwarderCreationData)(this.transactionData);
            this.baseAddress = baseAddress;
            this.salt = addressCreationSalt;
            this.feeAddress = feeAddress;
            const forwarderImplementationAddress = this._coinConfig.network
                .forwarderImplementationAddress;
            this.initCode = (0, abstract_eth_1.getProxyInitcode)(forwarderImplementationAddress);
            // Set sender address
            if (signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.AddressInitializationTransaction = AddressInitializationTransaction;
