// @bitgo/sdk-coin-vet/dist/src/lib/transaction/claimRewards.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimRewardsTransaction = void 0;
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const ethereumjs_util_1 = require("ethereumjs-util");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const transaction_1 = require("./transaction");
const constants_1 = require("../constants");
const utils_1 = __importDefault(require("../utils"));
class ClaimRewardsTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.StakingClaim;
    }
    get stakingContractAddress() {
        return this._stakingContractAddress;
    }
    set stakingContractAddress(address) {
        this._stakingContractAddress = address;
    }
    get tokenId() {
        return this._tokenId;
    }
    set tokenId(tokenId) {
        this._tokenId = tokenId;
    }
    /** @inheritdoc */
    async build() {
        this.buildClauses();
        await this.buildRawTransaction();
        this.generateTxnIdAndSetSender();
        this.loadInputsAndOutputs();
    }
    get clauses() {
        return this._clauses;
    }
    set clauses(clauses) {
        this._clauses = clauses;
    }
    get recipients() {
        return this._recipients;
    }
    set recipients(recipients) {
        this._recipients = recipients;
    }
    /** @inheritdoc */
    buildClauses() {
        if (!this.stakingContractAddress) {
            throw new Error('Staking contract address is not set');
        }
        utils_1.default.validateStakingContractAddress(this.stakingContractAddress, this._coinConfig);
        if (this.tokenId === undefined || this.tokenId === null) {
            throw new Error('Token ID is not set');
        }
        const data = this.encodeClaimRewardsMethod(this.tokenId);
        this._transactionData = data;
        // Create the clause for claim rewards
        this._clauses = [
            {
                to: this.stakingContractAddress,
                value: '0x0',
                data: this._transactionData,
            },
        ];
        // Set recipients as empty since claim rewards doesn't send value
        this.recipients = [];
    }
    /**
     * Encode the claim rewards method call data
     */
    encodeClaimRewardsMethod(tokenId) {
        const methodName = 'claimRewards';
        const types = ['uint256'];
        const params = [tokenId];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    /** @inheritdoc */
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            sender: this.sender,
            feePayer: this.feePayerAddress,
            recipients: this.recipients,
            tokenId: this.tokenId,
            stakingContractAddress: this.stakingContractAddress,
        };
        return json;
    }
    /** @inheritdoc */
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            if (body.clauses.length === 1) {
                const clause = body.clauses[0];
                if (clause.data && clause.data.startsWith(constants_1.CLAIM_STAKING_REWARDS_METHOD_ID)) {
                    // claimRewards should go to STARGATE_DELEGATION_ADDRESS
                    this.tokenId = utils_1.default.decodeClaimRewardsData(clause.data);
                    this.stakingContractAddress = clause.to || '0x0';
                }
            }
            // Set recipients as empty for claim rewards
            this.recipients = [];
            this.loadInputsAndOutputs();
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.ClaimRewardsTransaction = ClaimRewardsTransaction;
