// @bitgo/sdk-coin-vet/dist/src/lib/transaction/delegateClauseTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegateClauseTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const transaction_1 = require("./transaction");
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const utils_1 = __importDefault(require("../utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const ethereumjs_util_1 = require("ethereumjs-util");
const constants_1 = require("../constants");
class DelegateClauseTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.StakingDelegate;
    }
    get validator() {
        return this._validator;
    }
    set validator(address) {
        this._validator = address;
    }
    get stakingContractAddress() {
        return this._stakingContractAddress;
    }
    set stakingContractAddress(address) {
        this._stakingContractAddress = address;
    }
    get tokenId() {
        return this._tokenId;
    }
    set tokenId(tokenId) {
        this._tokenId = tokenId;
    }
    buildClauses() {
        if (!this.stakingContractAddress) {
            throw new Error('Staking contract address is not set');
        }
        utils_1.default.validateStakingContractAddress(this.stakingContractAddress, this._coinConfig);
        if (this.tokenId === undefined || this.tokenId === null) {
            throw new Error('Token ID is not set');
        }
        if (this.validator === undefined || this.validator === null) {
            throw new Error('Validator address is not set');
        }
        const data = this.getDelegateData(this.tokenId, this.validator);
        this._transactionData = data;
        // Create the clause for delegation
        this._clauses = [
            {
                to: this.stakingContractAddress,
                value: constants_1.ZERO_VALUE_AMOUNT,
                data: this._transactionData,
            },
        ];
        // Set recipients based on the clauses
        this._recipients = [
            {
                address: this.stakingContractAddress,
                amount: constants_1.ZERO_VALUE_AMOUNT,
            },
        ];
    }
    /**
     * Encodes delegation transaction data using ethereumjs-abi for delegate method
     *
     * @param {number} tokenId - The Token ID for delegation
     * @returns {string} - The encoded transaction data
     */
    getDelegateData(tokenId, validatorAddress) {
        const methodName = 'delegate';
        const types = ['uint256', 'address'];
        const params = [new ethereumjs_util_1.BN(tokenId), validatorAddress];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            data: this.transactionData,
            value: constants_1.ZERO_VALUE_AMOUNT,
            sender: this.sender,
            to: this.stakingContractAddress,
            stakingContractAddress: this.stakingContractAddress,
            amountToStake: constants_1.ZERO_VALUE_AMOUNT,
            tokenId: this.tokenId,
            validatorAddress: this.validator,
        };
        return json;
    }
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set delegation-specific properties
            if (body.clauses.length > 0) {
                const clause = body.clauses[0];
                if (clause.to) {
                    this.stakingContractAddress = clause.to;
                }
                if (clause.data) {
                    this.transactionData = clause.data;
                    const decoded = utils_1.default.decodeDelegateClauseData(clause.data);
                    this.tokenId = decoded.tokenId;
                    this.validator = decoded.validator;
                }
            }
            // Set recipients from clauses
            this.recipients = body.clauses.map((clause) => ({
                address: (clause.to || '0x0').toString().toLowerCase(),
                amount: new bignumber_js_1.default(clause.value || 0).toString(),
            }));
            this.loadInputsAndOutputs();
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.DelegateClauseTransaction = DelegateClauseTransaction;
