// @bitgo/sdk-coin-vet/dist/src/lib/transaction/exitDelegation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExitDelegationTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const transaction_1 = require("./transaction");
const constants_1 = require("../constants");
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const ethereumjs_util_1 = require("ethereumjs-util");
const utils_1 = __importDefault(require("../utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class ExitDelegationTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.StakingUnlock;
    }
    get tokenId() {
        return this._tokenId;
    }
    set tokenId(id) {
        this._tokenId = id;
    }
    get stakingContractAddress() {
        return this._stakingContractAddress;
    }
    set stakingContractAddress(address) {
        this._stakingContractAddress = address;
    }
    /** @inheritdoc */
    async build() {
        this.buildClauses();
        await this.buildRawTransaction();
        this.generateTxnIdAndSetSender();
        this.loadInputsAndOutputs();
    }
    /** @inheritdoc */
    buildClauses() {
        if (!this._stakingContractAddress || !this._tokenId) {
            throw new sdk_core_1.InvalidTransactionError('Missing required unstaking parameters');
        }
        utils_1.default.validateStakingContractAddress(this._stakingContractAddress, this._coinConfig);
        this._clauses = [
            {
                to: this._stakingContractAddress,
                value: '0x0',
                data: this._transactionData || this.getExitDelegationData(),
            },
        ];
        this._recipients = [
            {
                address: this._stakingContractAddress,
                amount: '0',
            },
        ];
    }
    /**
     * Generates the transaction data for unstaking by encoding the exitDelegation method call.
     *
     * @private
     * @returns {string} The encoded transaction data as a hex string
     */
    getExitDelegationData() {
        const methodName = 'requestDelegationExit';
        const types = ['uint256'];
        const params = [this._tokenId];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    /** @inheritdoc */
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            data: this.transactionData || this.getExitDelegationData(),
            value: '0',
            sender: this.sender,
            to: this.stakingContractAddress,
            tokenId: this.tokenId,
        };
        return json;
    }
    /** @inheritdoc */
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set data from clauses
            this.stakingContractAddress = body.clauses[0]?.to || '0x0';
            this.transactionData = body.clauses[0]?.data || '0x0';
            this.type = sdk_core_1.TransactionType.StakingUnlock;
            // Extract tokenId from transaction data
            if (this.transactionData.startsWith(constants_1.EXIT_DELEGATION_METHOD_ID)) {
                this.tokenId = utils_1.default.decodeExitDelegationData(this.transactionData);
            }
            this.recipients = body.clauses.map((clause) => ({
                address: (clause.to || '0x0').toString().toLowerCase(),
                amount: new bignumber_js_1.default(clause.value || 0).toString(),
            }));
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            this.loadInputsAndOutputs();
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.ExitDelegationTransaction = ExitDelegationTransaction;
