// @bitgo/sdk-coin-vet/dist/src/lib/transaction/flushTokenTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlushTokenTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const transaction_1 = require("./transaction");
class FlushTokenTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.FlushTokens;
    }
    get forwarderVersion() {
        return this._forwarderVersion;
    }
    set forwarderVersion(forwarderVersion) {
        this._forwarderVersion = forwarderVersion;
    }
    get tokenAddress() {
        return this._tokenAddress;
    }
    set tokenAddress(address) {
        this._tokenAddress = address;
    }
    /** @inheritdoc */
    buildClauses() {
        this._clauses = [
            {
                to: this._contract,
                value: '0x0',
                data: this._transactionData,
            },
        ];
    }
    /** @inheritdoc */
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            data: this.transactionData,
            value: '0',
            sender: this.sender,
            to: this.contract,
            tokenAddress: this.tokenAddress,
        };
        return json;
    }
    /** @inheritdoc */
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set data from clauses
            this.contract = body.clauses[0]?.to || '0x0';
            this.transactionData = body.clauses[0]?.data || '0x0';
            this.type = sdk_core_1.TransactionType.FlushTokens;
            const { tokenAddress } = (0, abstract_eth_1.decodeFlushTokensData)(this.transactionData, this.contract);
            this.tokenAddress = tokenAddress;
            // Set sender address
            if (signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.FlushTokenTransaction = FlushTokenTransaction;
