// @bitgo/sdk-coin-vet/dist/src/lib/transaction/nftTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NFTTransaction = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@vechain/sdk-core");
const sdk_core_2 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
const transaction_1 = require("./transaction");
class NFTTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_2.TransactionType.SendNFT;
    }
    get nftCollectionId() {
        return this._nftCollectionId;
    }
    set nftCollectionId(nftCollectionId) {
        this._nftCollectionId = nftCollectionId;
    }
    get tokenId() {
        return this._tokenId;
    }
    set tokenId(tokenId) {
        this._tokenId = tokenId;
    }
    buildClauses() {
        if (!this.nftCollectionId) {
            throw new Error('NFT collection id is not set');
        }
        if (!this.sender) {
            throw new Error('Sender address is not set');
        }
        if (!this.tokenId) {
            throw new Error('Token id is not set');
        }
        this.clauses = this.recipients.map((recipient) => {
            const data = utils_1.default.getTransferNFTData(this.sender, recipient.address, this.tokenId);
            return {
                to: this.nftCollectionId,
                value: '0x0',
                data,
            };
        });
    }
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            recipients: this.recipients,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            sender: this.sender,
            feePayer: this.feePayerAddress,
            nftCollectionId: this.nftCollectionId,
            tokenId: this.tokenId,
        };
        return json;
    }
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_2.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = body.chainTag;
            this.blockRef = body.blockRef;
            this.expiration = body.expiration;
            this.clauses = body.clauses;
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = Number(body.gas);
            this.dependsOn = body.dependsOn;
            this.nonce = String(body.nonce);
            // Set recipients from clauses
            (0, assert_1.default)(body.clauses[0].to, 'nft collection id(contract address) address not found in the clauses');
            (0, assert_1.default)(body.clauses.length === 1, 'NFT transaction should have exactly one clause');
            this.nftCollectionId = body.clauses[0].to;
            const decodedData = utils_1.default.decodeTransferNFTData(body.clauses[0].data);
            this.recipients = decodedData.recipients;
            this.tokenId = decodedData.tokenId;
            this.sender = decodedData.sender;
            this.loadInputsAndOutputs();
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_1.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_1.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_1.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_2.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.NFTTransaction = NFTTransaction;
