// @bitgo/sdk-coin-vet/dist/src/lib/transaction/stakeClauseTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakeClauseTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const transaction_1 = require("./transaction");
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const utils_1 = __importDefault(require("../utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const ethereumjs_util_1 = require("ethereumjs-util");
class StakeClauseTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.StakingActivate;
    }
    get stakingContractAddress() {
        return this._stakingContractAddress;
    }
    set stakingContractAddress(address) {
        this._stakingContractAddress = address;
    }
    get levelId() {
        return this._levelId;
    }
    set levelId(levelId) {
        this._levelId = levelId;
    }
    get amountToStake() {
        return this._amountToStake;
    }
    set amountToStake(amount) {
        this._amountToStake = amount;
    }
    buildClauses() {
        if (!this.stakingContractAddress) {
            throw new Error('Staking contract address is not set');
        }
        utils_1.default.validateStakingContractAddress(this.stakingContractAddress, this._coinConfig);
        if (this.levelId === undefined || this.levelId === null) {
            throw new Error('Level ID is not set');
        }
        if (!this.amountToStake) {
            throw new Error('Amount to stake is not set');
        }
        const data = this.getStakingData(this.levelId);
        this._transactionData = data;
        // Create the clause for staking
        this._clauses = [
            {
                to: this.stakingContractAddress,
                value: this.amountToStake,
                data: this._transactionData,
            },
        ];
        // Set recipients based on the clauses
        this._recipients = [
            {
                address: this.stakingContractAddress,
                amount: this.amountToStake,
            },
        ];
    }
    /**
     * Encodes staking transaction data using ethereumjs-abi for stake method
     *
     * @param {number} levelId - The level ID for staking
     * @returns {string} - The encoded transaction data
     */
    getStakingData(levelId) {
        const methodName = 'stake';
        const types = ['uint8'];
        const params = [levelId];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            data: this.transactionData,
            value: this.amountToStake,
            sender: this.sender,
            to: this.stakingContractAddress,
            stakingContractAddress: this.stakingContractAddress,
            amountToStake: this.amountToStake,
            levelId: this.levelId,
        };
        return json;
    }
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set staking-specific properties
            if (body.clauses.length > 0) {
                const clause = body.clauses[0];
                if (clause.to) {
                    this.stakingContractAddress = clause.to;
                }
                if (clause.value) {
                    this.amountToStake = new bignumber_js_1.default(clause.value).toFixed();
                }
                if (clause.data) {
                    this.transactionData = clause.data;
                    const decoded = utils_1.default.decodeStakeClauseData(clause.data);
                    this.levelId = decoded.levelId;
                }
            }
            // Set recipients from clauses
            this.recipients = body.clauses.map((clause) => ({
                address: (clause.to || '0x0').toString().toLowerCase(),
                amount: new bignumber_js_1.default(clause.value || 0).toString(),
            }));
            this.loadInputsAndOutputs();
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.StakeClauseTransaction = StakeClauseTransaction;
