// @bitgo/sdk-coin-vet/dist/src/lib/transaction/tokenTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransaction = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@vechain/sdk-core");
const sdk_core_2 = require("@bitgo/sdk-core");
const transaction_1 = require("./transaction");
const utils_1 = __importDefault(require("../utils"));
class TokenTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_2.TransactionType.Send;
    }
    get tokenAddress() {
        return this._tokenAddress;
    }
    set tokenAddress(address) {
        this._tokenAddress = address;
    }
    buildClauses() {
        if (!this.tokenAddress) {
            throw new Error('Token address is not set');
        }
        this.clauses = this.recipients.map((recipient) => {
            const data = utils_1.default.getTransferTokenData(recipient.address, String(recipient.amount));
            return {
                to: this.tokenAddress,
                value: '0x0',
                data,
            };
        });
    }
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            recipients: this.recipients,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            sender: this.sender,
            feePayer: this.feePayerAddress,
            tokenAddress: this.tokenAddress,
        };
        return json;
    }
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_2.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = body.chainTag;
            this.blockRef = body.blockRef;
            this.expiration = body.expiration;
            this.clauses = body.clauses;
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = Number(body.gas);
            this.dependsOn = body.dependsOn;
            this.nonce = String(body.nonce);
            // Set recipients from clauses
            (0, assert_1.default)(body.clauses[0].to, 'token address not found in the clauses');
            this.tokenAddress = body.clauses[0].to;
            this.recipients = body.clauses.map((clause) => utils_1.default.decodeTransferTokenData(clause.data));
            this.loadInputsAndOutputs();
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_1.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_1.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_1.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_2.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.TokenTransaction = TokenTransaction;
