// @bitgo/sdk-coin-vet/dist/src/lib/transaction/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const nc_utils = __importStar(require("@noble/curves/abstract/utils"));
const utils_1 = __importDefault(require("../utils"));
const gasPrice = 1e13;
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.Send;
        this._chainTag = 0x27; // Initialize to 39 (0x27)
        this._blockRef = '0x0';
        this._expiration = 64;
        this._clauses = [];
        this._gasPriceCoef = 128;
        this._gas = 0;
        this._dependsOn = null;
        this._nonce = String(0);
        this._recipients = [];
        this._senderSignature = null;
        this._feePayerSignature = null;
        this._isRecovery = false;
    }
    get id() {
        this.generateTxnIdAndSetSender();
        return this._id ?? 'UNAVAILABLE';
    }
    get rawTransaction() {
        return this._rawTransaction;
    }
    set rawTransaction(rawTx) {
        this._rawTransaction = rawTx;
    }
    get type() {
        return this._type;
    }
    set type(type) {
        this._type = type;
    }
    get sender() {
        return this._sender;
    }
    set sender(address) {
        this._sender = address;
    }
    get senderSignature() {
        if (this._senderSignature) {
            return new Uint8Array(this._senderSignature);
        }
        else {
            return undefined;
        }
    }
    set senderSignature(sig) {
        this._senderSignature = sig;
    }
    get feePayerAddress() {
        return this._feePayerAddress;
    }
    set feePayerAddress(address) {
        this._feePayerAddress = address;
    }
    get feePayerSignature() {
        if (this._feePayerSignature) {
            return new Uint8Array(this._feePayerSignature);
        }
        else {
            return undefined;
        }
    }
    set feePayerSignature(sig) {
        this._feePayerSignature = sig;
    }
    get chainTag() {
        return this._chainTag;
    }
    set chainTag(tag) {
        this._chainTag = tag;
    }
    get blockRef() {
        return this._blockRef;
    }
    set blockRef(ref) {
        this._blockRef = ref;
    }
    get expiration() {
        return this._expiration;
    }
    set expiration(exp) {
        this._expiration = exp;
    }
    get recipients() {
        return this._recipients;
    }
    set recipients(recipients) {
        this._recipients = recipients;
    }
    get clauses() {
        return this._clauses;
    }
    set clauses(clauses) {
        this._clauses = clauses;
    }
    get gasPriceCoef() {
        return this._gasPriceCoef;
    }
    set gasPriceCoef(coef) {
        this._gasPriceCoef = coef;
    }
    get gas() {
        return this._gas;
    }
    set gas(g) {
        this._gas = g;
    }
    get dependsOn() {
        return this._dependsOn;
    }
    set dependsOn(dep) {
        this._dependsOn = dep;
    }
    get nonce() {
        return this._nonce;
    }
    set nonce(n) {
        this._nonce = n;
    }
    get contract() {
        return this._contract;
    }
    set contract(address) {
        this._contract = address;
    }
    get transactionData() {
        return this._transactionData;
    }
    set transactionData(transactionData) {
        this._transactionData = transactionData;
    }
    get isRecovery() {
        return this._isRecovery;
    }
    set isRecovery(isRecovery) {
        this._isRecovery = isRecovery;
    }
    /**
     * Get all signatures associated with this transaction
     * Required by BaseTransaction
     */
    get signature() {
        const sigs = [];
        if (this._senderSignature) {
            sigs.push(Buffer.from(this._senderSignature).toString('hex'));
        }
        if (this._feePayerSignature) {
            sigs.push(Buffer.from(this._feePayerSignature).toString('hex'));
        }
        return sigs;
    }
    /**
     * Set signatures for this transaction
     * Required by BaseTransaction
     * Note: This is mainly for compatibility with the base class.
     * Prefer using senderSignature and feePayerSignature directly.
     */
    set signature(sigs) {
        if (sigs.length > 0) {
            this._senderSignature = Buffer.from(sigs[0], 'hex');
        }
        if (sigs.length > 1) {
            this._feePayerSignature = Buffer.from(sigs[1], 'hex');
        }
    }
    /** @inheritdoc */
    canSign() {
        return true;
    }
    /**
     * Check if this is a fee delegated transaction
     */
    isFeeDelegated() {
        return !!this._feePayerAddress;
    }
    addRecipients(clauses) {
        this._recipients = clauses.map((clause) => ({
            address: (clause.to || '0x0').toString().toLowerCase(),
            amount: (clause.value || '0').toString(),
        }));
    }
    buildClauses() {
        this._clauses = this._recipients.map((recipient) => ({
            to: recipient.address,
            value: recipient.amount,
            data: '0x',
        }));
    }
    /**
     * Populates the transaction object with data from a deserialized VeChain transaction
     * @param signedTx - The deserialized VeChain transaction containing body, signatures, and other transaction data
     * @throws {InvalidTransactionError} When the transaction is invalid or missing required data
     */
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set recipients from clauses
            this.recipients = body.clauses.map((clause) => ({
                address: (clause.to || '0x0').toString().toLowerCase(),
                amount: new bignumber_js_1.default(clause.value).toFixed(),
            }));
            this.loadInputsAndOutputs();
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
    addSenderSignature(signature) {
        this.senderSignature = signature;
    }
    addFeePayerSignature(signature) {
        this.feePayerSignature = signature;
    }
    async build() {
        this.buildClauses();
        await this.buildRawTransaction();
        this.generateTxnIdAndSetSender();
        this.loadInputsAndOutputs();
    }
    /**
     * Sets the transaction ID from the raw transaction if it is signed
     * @protected
     */
    generateTxnIdAndSetSender() {
        // Check if we have a raw transaction
        if (!this.rawTransaction) {
            return;
        }
        if (!this.senderSignature) {
            return;
        }
        else {
            const halfSignedTransaction = sdk_core_2.Transaction.of(this.rawTransaction.body, this.senderSignature);
            if (halfSignedTransaction.signature) {
                this._rawTransaction = halfSignedTransaction;
                this._sender = halfSignedTransaction.origin.toString().toLowerCase();
                if (this.isRecovery) {
                    this._id = halfSignedTransaction.id.toString();
                }
            }
            else {
                return;
            }
            if (this.feePayerSignature) {
                const fullSignedTransaction = sdk_core_2.Transaction.of(halfSignedTransaction.body, nc_utils.concatBytes(
                // Drop any previous gas payer signature.
                halfSignedTransaction.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH), this.feePayerSignature));
                if (fullSignedTransaction.signature) {
                    this._rawTransaction = fullSignedTransaction;
                    this._id = fullSignedTransaction.id.toString();
                }
                else {
                    return;
                }
            }
        }
    }
    async buildRawTransaction() {
        const transactionBody = {
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            clauses: this.clauses,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: null,
            nonce: this.nonce,
        };
        if ((this.type === sdk_core_1.TransactionType.Send && !this.isRecovery) ||
            this.type === sdk_core_1.TransactionType.SendToken ||
            this.type === sdk_core_1.TransactionType.SendNFT ||
            this.type === sdk_core_1.TransactionType.ContractCall ||
            this.type === sdk_core_1.TransactionType.StakingActivate ||
            this.type === sdk_core_1.TransactionType.StakingDelegate ||
            this.type === sdk_core_1.TransactionType.StakingUnlock ||
            this.type === sdk_core_1.TransactionType.StakingWithdraw ||
            this.type === sdk_core_1.TransactionType.StakingClaim ||
            this.type === sdk_core_1.TransactionType.StakingLock ||
            this.type === sdk_core_1.TransactionType.StakingAdd) {
            transactionBody.reserved = {
                features: 1, // mark transaction as delegated i.e. will use gas payer
            };
        }
        this.rawTransaction = sdk_core_2.Transaction.of(transactionBody);
    }
    loadInputsAndOutputs() {
        const totalAmount = this._recipients.reduce((accumulator, current) => accumulator.plus(current.amount), new bignumber_js_1.default('0'));
        this._inputs = [
            {
                address: this.sender,
                value: totalAmount.toString(),
                coin: this._coinConfig.name,
            },
        ];
        this._outputs = this._recipients.map((recipient) => {
            return {
                address: recipient.address,
                value: recipient.amount,
                coin: this._coinConfig.name,
            };
        });
    }
    fromRawTransaction(rawTransaction) {
        let signedTxn;
        try {
            signedTxn = utils_1.default.deserializeTransaction(rawTransaction);
        }
        catch (e) {
            throw new Error('invalid raw transaction');
        }
        this.fromDeserializedSignedTransaction(signedTxn);
    }
    /** @inheritdoc */
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            recipients: this.recipients,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            sender: this.sender,
            feePayer: this.feePayerAddress,
        };
        return json;
    }
    getFee() {
        return new bignumber_js_1.default(this.gas)
            .multipliedBy(gasPrice)
            .multipliedBy(new bignumber_js_1.default(this.gasPriceCoef).dividedBy(255).plus(1))
            .toString()
            .slice(0, 18);
    }
    get signablePayload() {
        return Buffer.from(this.rawTransaction.encoded);
    }
    /** @inheritdoc */
    toBroadcastFormat() {
        if (!this.rawTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty Transaction');
        }
        return this.serialize();
    }
    serialize() {
        if (!this.senderSignature) {
            return sdk_core_2.HexUInt.of(this.rawTransaction.encoded).toString();
        }
        else {
            if (!this.feePayerSignature) {
                const senderSignedTransaction = sdk_core_2.Transaction.of(this.rawTransaction.body, this.senderSignature);
                return sdk_core_2.HexUInt.of(senderSignedTransaction.encoded).toString();
            }
            else {
                const senderSignedTransaction = sdk_core_2.Transaction.of(this.rawTransaction.body, this.senderSignature);
                if (senderSignedTransaction.signature) {
                    const fullSignedTransaction = sdk_core_2.Transaction.of(senderSignedTransaction.body, nc_utils.concatBytes(senderSignedTransaction.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH), this.feePayerSignature));
                    return sdk_core_2.HexUInt.of(fullSignedTransaction.encoded).toString();
                }
                else {
                    throw new sdk_core_1.InvalidTransactionError('Transaction is not signed properly');
                }
            }
        }
    }
    static deserializeTransaction(rawTx) {
        try {
            if (!rawTx) {
                throw new sdk_core_1.InvalidTransactionError('Raw transaction string cannot be empty');
            }
            return utils_1.default.deserializeTransaction(rawTx);
        }
        catch (e) {
            if (e instanceof sdk_core_1.InvalidTransactionError) {
                throw e;
            }
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
    explainTransaction() {
        const displayOrder = [
            'id',
            'outputs',
            'outputAmount',
            'changeOutputs',
            'changeAmount',
            'fee',
            'withdrawAmount',
            'sender',
            'type',
        ];
        const outputs = this._recipients;
        const outputAmount = outputs
            .reduce((accumulator, current) => accumulator.plus(current.amount), new bignumber_js_1.default('0'))
            .toString();
        return {
            displayOrder,
            id: this.id,
            outputs,
            outputAmount,
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: this.getFee() },
            sender: this.sender,
            type: this.type,
        };
    }
}
exports.Transaction = Transaction;
