// @bitgo/sdk-coin-vet/dist/src/lib/transaction/validatorRegistrationTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorRegistrationTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_core_2 = require("@vechain/sdk-core");
const transaction_1 = require("./transaction");
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const utils_1 = __importDefault(require("../utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const ethereumjs_util_1 = require("ethereumjs-util");
class ValidatorRegistrationTransaction extends transaction_1.Transaction {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.StakingLock;
    }
    get validator() {
        return this._validator;
    }
    set validator(address) {
        this._validator = address;
    }
    get stakingPeriod() {
        return this._stakingPeriod;
    }
    set stakingPeriod(period) {
        this._stakingPeriod = period;
    }
    get amountToStake() {
        return this._amountToStake;
    }
    set amountToStake(amount) {
        this._amountToStake = amount;
    }
    get stakingContractAddress() {
        return this._stakingContractAddress;
    }
    set stakingContractAddress(address) {
        this._stakingContractAddress = address;
    }
    buildClauses() {
        if (!this.stakingContractAddress) {
            throw new Error('Staking contract address is not set');
        }
        if (!this.validator) {
            throw new Error('Validator address is not set');
        }
        if (!this.stakingPeriod) {
            throw new Error('Staking period is not set');
        }
        utils_1.default.validateContractAddressForValidatorRegistration(this.stakingContractAddress, this._coinConfig);
        const addValidationData = this.getAddValidationClauseData(this.validator, this.stakingPeriod);
        this._transactionData = addValidationData;
        // Create the clause for delegation
        this._clauses = [
            {
                to: this.stakingContractAddress,
                value: this.amountToStake,
                data: addValidationData,
            },
        ];
        // Set recipients based on the clauses
        this._recipients = [
            {
                address: this.stakingContractAddress,
                amount: this.amountToStake,
            },
        ];
    }
    /**
     * Encodes addValidation transaction data using ethereumjs-abi for addValidation method
     * @param {string} validator - address of the validator
     * @param {number} period - staking period, denoted in blocks, that the Validator commits to hard
  locking their VET into the built-in staker contract. Allowed values are 60480 (7 days),
  129600 (15 days) or 259200 (30 days)
     * @returns {string} - The encoded transaction data
     */
    getAddValidationClauseData(validator, period) {
        const methodName = 'addValidation';
        const types = ['address', 'uint32'];
        const params = [validator, new ethereumjs_util_1.BN(period)];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    toJson() {
        const json = {
            id: this.id,
            chainTag: this.chainTag,
            blockRef: this.blockRef,
            expiration: this.expiration,
            gasPriceCoef: this.gasPriceCoef,
            gas: this.gas,
            dependsOn: this.dependsOn,
            nonce: this.nonce,
            data: this.transactionData,
            value: this.amountToStake,
            sender: this.sender,
            to: this.stakingContractAddress,
            stakingContractAddress: this.stakingContractAddress,
            amountToStake: this.amountToStake,
            validatorAddress: this.validator,
            stakingPeriod: this.stakingPeriod,
        };
        return json;
    }
    fromDeserializedSignedTransaction(signedTx) {
        try {
            if (!signedTx || !signedTx.body) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: missing transaction body');
            }
            // Store the raw transaction
            this.rawTransaction = signedTx;
            // Set transaction body properties
            const body = signedTx.body;
            this.chainTag = typeof body.chainTag === 'number' ? body.chainTag : 0;
            this.blockRef = body.blockRef || '0x0';
            this.expiration = typeof body.expiration === 'number' ? body.expiration : 64;
            this.clauses = body.clauses || [];
            this.gasPriceCoef = typeof body.gasPriceCoef === 'number' ? body.gasPriceCoef : 128;
            this.gas = typeof body.gas === 'number' ? body.gas : Number(body.gas) || 0;
            this.dependsOn = body.dependsOn || null;
            this.nonce = String(body.nonce);
            // Set validator registration-specific properties
            if (body.clauses.length > 0) {
                // Get the addValidation clause
                const addValidationClause = body.clauses[0];
                if (addValidationClause.to) {
                    this.stakingContractAddress = addValidationClause.to;
                }
                if (addValidationClause.value) {
                    this.amountToStake = new bignumber_js_1.default(addValidationClause.value).toFixed();
                }
                // Extract validator and period from addValidation data
                if (addValidationClause.data) {
                    this.transactionData = addValidationClause.data;
                    const decoded = utils_1.default.decodeAddValidationData(addValidationClause.data);
                    this.validator = decoded.validator;
                    this.stakingPeriod = decoded.period;
                }
            }
            // Set recipients from clauses
            this.recipients = body.clauses.map((clause) => ({
                address: (clause.to || '0x0').toString().toLowerCase(),
                amount: new bignumber_js_1.default(clause.value || 0).toString(),
            }));
            this.loadInputsAndOutputs();
            // Set sender address
            if (signedTx.signature && signedTx.origin) {
                this.sender = signedTx.origin.toString().toLowerCase();
            }
            // Set signatures if present
            if (signedTx.signature) {
                // First signature is sender's signature
                this.senderSignature = Buffer.from(signedTx.signature.slice(0, sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                // If there's additional signature data, it's the fee payer's signature
                if (signedTx.signature.length > sdk_core_2.Secp256k1.SIGNATURE_LENGTH) {
                    this.feePayerSignature = Buffer.from(signedTx.signature.slice(sdk_core_2.Secp256k1.SIGNATURE_LENGTH));
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to deserialize transaction: ${e.message}`);
        }
    }
}
exports.ValidatorRegistrationTransaction = ValidatorRegistrationTransaction;
