// @bitgo/sdk-coin-vet/dist/test/unit/vet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sinon_1 = __importDefault(require("sinon"));
const assert_1 = __importDefault(require("assert"));
const lodash_1 = __importDefault(require("lodash"));
const should_1 = __importDefault(require("should"));
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const statics_1 = require("@bitgo/statics");
const src_1 = require("../../src");
const testData = __importStar(require("../resources/vet"));
describe('Vechain', function () {
    let bitgo;
    let basecoin;
    let newTxPrebuild;
    let newTxParams;
    const txPreBuild = {
        txHex: testData.SPONSORED_TRANSACTION,
        txInfo: {},
    };
    const txParams = {
        recipients: testData.recipients,
    };
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('vet', src_1.Vet.createInstance);
        bitgo.safeRegister('tvet', src_1.Tvet.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('tvet');
        newTxPrebuild = () => {
            return lodash_1.default.cloneDeep(txPreBuild);
        };
        newTxParams = () => {
            return lodash_1.default.cloneDeep(txParams);
        };
    });
    it('should return the right info', function () {
        const vet = bitgo.coin('vet');
        const tvet = bitgo.coin('tvet');
        const vetStatics = statics_1.coins.get('vet');
        const tvetStatics = statics_1.coins.get('tvet');
        vet.getChain().should.equal('vet');
        vet.getFamily().should.equal('vet');
        vet.getFullName().should.equal('VeChain');
        vet.getBaseFactor().should.equal(1e18);
        tvet.getChain().should.equal('tvet');
        tvet.getFamily().should.equal('vet');
        tvet.getFullName().should.equal('Testnet VeChain');
        tvet.getBaseFactor().should.equal(1e18);
        vetStatics.gasTankLowBalanceAlertFactor.should.equal(80);
        tvetStatics.gasTankLowBalanceAlertFactor.should.equal(80);
        vetStatics.gasTankMinBalanceRecommendationFactor.should.equal(200);
        tvetStatics.gasTankMinBalanceRecommendationFactor.should.equal(200);
    });
    it('should validate address', function () {
        const vet = bitgo.coin('vet');
        vet.isValidAddress('wrongaddress').should.false();
        vet.isValidAddress('25bcb8855effa9f12a23c2f7f34f2d92b5841f19').should.false();
        vet.isValidAddress('0x7Ca00e3bC8a836026C2917C6c7c6D049E52099dd').should.true();
        vet.isValidAddress('0x690fFcefa92876C772E85d4B5963807C2152e08d').should.true();
        vet.isValidAddress('0xe59F1cea4e0FEf511e3d0f4EEc44ADf19C4cbeEC').should.true();
    });
    it('is valid pub', function () {
        // with 0x prefix
        basecoin.isValidPub('0x9b4e96086d111500259f9b38680b0509a405c1904da18976455a20c691d3bb07').should.equal(false);
        // without 0x prefix
        basecoin.isValidPub('027c10f30d5c874cee3d193a321c82926d905998cb4852880935da4ee1820bd7d5').should.equal(true);
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify transaction when recipients amount are numbers', async function () {
            const txPrebuild = newTxPrebuild();
            const txParamsWithNumberAmounts = newTxParams();
            txParamsWithNumberAmounts.recipients = txParamsWithNumberAmounts.recipients.map(({ address, amount }) => {
                return { address, amount: Number(amount) };
            });
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams: txParamsWithNumberAmounts,
                txPrebuild,
                verification,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = newTxParams();
            txParams.recipients = undefined;
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
        it('should build correct nft transfer data', function () {
            const data = basecoin.buildNftTransferData({
                fromAddress: testData.addresses.validAddresses[0],
                recipientAddress: testData.addresses.validAddresses[1],
                type: 'ERC721',
                tokenId: '1234',
                tokenContractAddress: testData.NFT_CONTRACT_ADDRESS,
            });
            data.should.deepEqual({
                tokenType: 'ERC721',
                tokenQuantity: '1',
                tokenContractAddress: '0x887d9102f0003f1724d8fd5d4fe95a11572fcd77',
                tokenId: '1234',
            });
        });
        it('should build throw invalid address error', function () {
            (0, should_1.default)(() => basecoin.buildNftTransferData({
                fromAddress: testData.addresses.validAddresses[0],
                recipientAddress: testData.addresses.invalidAddresses[0],
                type: 'ERC721',
                tokenId: '1234',
                tokenContractAddress: testData.NFT_CONTRACT_ADDRESS,
            })).throwError('Invalid recipient address');
            (0, should_1.default)(() => basecoin.buildNftTransferData({
                fromAddress: testData.addresses.invalidAddresses[0],
                recipientAddress: testData.addresses.validAddresses[0],
                type: 'ERC721',
                tokenId: '1234',
                tokenContractAddress: testData.NFT_CONTRACT_ADDRESS,
            })).throwError('Invalid from address');
        });
    });
    describe('Parse and Explain Transactions: ', () => {
        const transferInputsResponse = [
            {
                address: testData.addresses.validAddresses[0],
                amount: testData.AMOUNT.toString(),
            },
        ];
        const transferOutputsResponse = [
            {
                address: testData.recipients[0].address,
                amount: testData.recipients[0].amount,
            },
        ];
        it('should parse a transfer transaction', async function () {
            const parsedTransaction = await basecoin.parseTransaction({
                txHex: testData.SPONSORED_TRANSACTION,
            });
            parsedTransaction.should.deepEqual({
                inputs: transferInputsResponse,
                outputs: transferOutputsResponse,
            });
        });
        it('should explain a transfer transaction', async function () {
            const rawTx = newTxPrebuild().txHex;
            const transaction = new src_1.Transaction(statics_1.coins.get('tvet'));
            transaction.fromRawTransaction(rawTx);
            const explainedTx = transaction.explainTransaction();
            explainedTx.should.deepEqual({
                displayOrder: [
                    'id',
                    'outputs',
                    'outputAmount',
                    'changeOutputs',
                    'changeAmount',
                    'fee',
                    'withdrawAmount',
                    'sender',
                    'type',
                ],
                id: '0x6d842d5dc5d59d4e8f0a8ec2757b430d1f19c06766fbc5b3db5ebac8a067a439',
                outputs: [
                    {
                        address: '0xe59f1cea4e0fef511e3d0f4eec44adf19c4cbeec',
                        amount: '100000000000000000',
                    },
                ],
                outputAmount: '100000000000000000',
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: '315411764705882352' },
                sender: '0x7ca00e3bc8a836026c2917c6c7c6d049e52099dd',
                type: 0,
            });
        });
        it('should fail to explain a invalid raw transaction', async function () {
            const rawTx = 'invalidRawTransaction';
            const transaction = new src_1.Transaction(statics_1.coins.get('tvet'));
            await assert_1.default.rejects(async () => transaction.fromRawTransaction(rawTx), {
                message: 'invalid raw transaction',
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Vet.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await basecoin
                .parseTransaction({ txHex: testData.INVALID_TRANSACTION })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
    describe('address validation', () => {
        it('should return true when validating a well formatted address prefixed with 0x', async function () {
            const address = testData.addresses.validAddresses[0];
            basecoin.isValidAddress(address).should.equal(true);
        });
        it('should return false when validating an incorrectly formatted', async function () {
            const address = 'wrongaddress';
            basecoin.isValidAddress(address).should.equal(false);
        });
    });
    describe('address verification', () => {
        const testData = {
            walletVersion: 6,
            commonKeychain: '02fad451e7d1a536897ded5f803a73cc7309a403cf43bb25bda494b9800efe32999b6b71b6159eccf1244492556b733f68733a23c184bf40bc37c52da5ad29e575',
            baseAddress: '0x8420429aa50b7f6ab196c4ce0dcf629fdbb821a1',
            feeAddress: '0x33f9d3172de8a88e47b399562e51fa9cde0f2511',
            depositAddress: '0x26b88c0a103d185792a9597580c2c5170d93f410',
            depositIndex: 3,
            forwarderVersion: 5,
        };
        it('should verify a valid TSS base address (wallet version 6)', async function () {
            const params = {
                address: testData.baseAddress,
                baseAddress: testData.baseAddress,
                keychains: [
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                ],
                index: 0,
                walletVersion: testData.walletVersion,
                coinSpecific: {
                    feeAddress: testData.feeAddress,
                    forwarderVersion: testData.forwarderVersion,
                },
            };
            const result = await basecoin.isWalletAddress(params);
            result.should.equal(true);
        });
        it('should fail to verify a invalid TSS base address (wallet version 6)', async function () {
            const params = {
                address: '0x8420429aa50b7f6ab196c4ce0dcf629fdbb821a0',
                baseAddress: testData.baseAddress,
                keychains: [
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                ],
                index: 0,
                walletVersion: testData.walletVersion,
                coinSpecific: {
                    feeAddress: testData.feeAddress,
                    forwarderVersion: testData.forwarderVersion,
                },
            };
            await basecoin.isWalletAddress(params).should.be.rejected();
        });
        it('should verify a valid TSS deposit address (wallet version 6)', async function () {
            const params = {
                address: testData.depositAddress,
                baseAddress: testData.baseAddress,
                keychains: [
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                ],
                index: testData.depositIndex,
                walletVersion: testData.walletVersion,
                coinSpecific: {
                    feeAddress: testData.feeAddress,
                    forwarderVersion: testData.forwarderVersion,
                },
            };
            const result = await basecoin.isWalletAddress(params);
            result.should.equal(true);
        });
        it('should fail to verify a invalid TSS deposit address (wallet version 6)', async function () {
            const params = {
                address: '0x26b88c0a103d185792a9597580c2c5170d93f411',
                baseAddress: testData.baseAddress,
                keychains: [
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                    { commonKeychain: testData.commonKeychain },
                ],
                index: testData.depositIndex,
                walletVersion: testData.walletVersion,
                coinSpecific: {
                    feeAddress: testData.feeAddress,
                    forwarderVersion: testData.forwarderVersion,
                },
            };
            await basecoin.isWalletAddress(params).should.be.rejected();
        });
    });
});
