// @bitgo/sdk-coin-sui/dist/test/unit/transactionBuilder/stakingBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const testData = __importStar(require("../../resources/sui"));
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const transaction_1 = require("../../../src/lib/transaction");
const iface_1 = require("../../../src/lib/iface");
const constants_1 = require("../../../src/lib/constants");
describe('Sui Staking Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsui');
    describe('Succeed', () => {
        it('should build a staking tx', async function () {
            const txBuilder = factory.getStakingBuilder();
            txBuilder.type(iface_1.SuiTransactionType.AddStake);
            txBuilder.sender(testData.sender.address);
            txBuilder.stake([testData.requestAddStake]);
            txBuilder.gasData(testData.gasData);
            const tx = await txBuilder.build();
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingAdd);
            tx.suiTransaction.gasData.payment.should.deepEqual(testData.coinsGasPayment);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: testData.STAKING_AMOUNT.toString(),
                coin: 'tsui',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: testData.requestAddStake.validatorAddress,
                value: testData.requestAddStake.amount.toString(),
                coin: 'tsui',
            });
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
            should_1.default.equal(rawTx, testData.ADD_STAKE);
        });
        it('should build a staking tx with more than 255 input objects', async function () {
            const numberOfPaymentObjects = 1000;
            const txBuilder = factory.getStakingBuilder();
            txBuilder.type(iface_1.SuiTransactionType.AddStake);
            txBuilder.sender(testData.sender.address);
            const gasData = {
                ...testData.gasData,
                payment: testData.generateObjects(numberOfPaymentObjects),
            };
            txBuilder.stake([testData.requestAddStake]);
            txBuilder.gasData(gasData);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.StakingAdd);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: testData.STAKING_AMOUNT.toString(),
                coin: 'tsui',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: testData.requestAddStake.validatorAddress,
                value: testData.requestAddStake.amount.toString(),
                coin: 'tsui',
            });
            tx.suiTransaction.gasData.owner.should.equal(gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(constants_1.MAX_GAS_OBJECTS - 1);
            const programmableTx = tx.suiTransaction.tx;
            // total objects - objects sent as gas payment
            // + (staked amount object + SUI system state object + validator address = 3)
            programmableTx.inputs.length.should.equal(numberOfPaymentObjects - (constants_1.MAX_GAS_OBJECTS - 1) + 3);
            programmableTx.transactions[0].kind.should.equal('MergeCoins');
            programmableTx.transactions[0].sources.length.should.equal(constants_1.MAX_COMMAND_ARGS - 1);
            programmableTx.transactions[1].kind.should.equal('MergeCoins');
            programmableTx.transactions[1].sources.length.should.equal(numberOfPaymentObjects - (constants_1.MAX_COMMAND_ARGS - 1) - (constants_1.MAX_GAS_OBJECTS - 1));
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
            const rebuilder = factory.from(rawTx);
            rebuilder.addSignature({ pub: testData.sender.publicKey }, Buffer.from(testData.sender.signatureHex));
            const rebuiltTx = await rebuilder.build();
            rebuiltTx.toBroadcastFormat().should.equal(rawTx);
            rebuiltTx.toJson().gasData.payment.length.should.equal(numberOfPaymentObjects);
        });
    });
    describe('Fail', () => {
        it('should fail for invalid sender', async function () {
            const builder = factory.getStakingBuilder();
            (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid or missing sender, got: randomString');
        });
        it('should fail for invalid gasData', function () {
            const builder = factory.getStakingBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasOwner)).throwError(`Invalid gas address ${testData.invalidGasOwner.owner}`);
        });
        it('should fail for invalid gasBudget', function () {
            const builder = factory.getStakingBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasBudget)).throwError('Invalid gas budget -1');
        });
        it('should fail for invalid gasPayment', function () {
            const builder = factory.getStakingBuilder();
            const invalidGasPayment = {
                ...testData.gasDataWithoutGasPayment,
                payment: [
                    {
                        objectId: '',
                        version: -1,
                        digest: '',
                    },
                ],
            };
            (0, should_1.default)(() => builder.gasData(invalidGasPayment)).throwError('Invalid payment, invalid or missing version');
        });
    });
});
