// @bitgo/sdk-coin-sui/dist/test/unit/transactionBuilder/tokenTransferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const testData = __importStar(require("../../resources/sui"));
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const transaction_1 = require("../../../src/lib/transaction");
const iface_1 = require("../../../src/lib/iface");
describe('Sui Token Transfer Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsui:deep');
    describe('Succeed', () => {
        it('should build a token transfer', async function () {
            const amount = 1000000000;
            const numberOfRecipients = 10;
            const numberOfInputObjects = 100;
            const numberOfGasPaymentObjects = 10;
            const txBuilder = factory.getTokenTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.TokenTransfer);
            txBuilder.sender(testData.sender.address);
            const recipients = new Array(numberOfRecipients).fill({
                address: testData.sender.address,
                amount: amount.toString(),
            });
            const gasData = {
                ...testData.gasData,
                payment: testData.generateObjects(numberOfGasPaymentObjects),
            };
            const inputObjects = testData.generateObjects(numberOfInputObjects);
            txBuilder.send(recipients);
            txBuilder.gasData(gasData);
            txBuilder.inputObjects(inputObjects);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.Send);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: (amount * numberOfRecipients).toString(),
                coin: 'tsui:deep',
            });
            tx.outputs.length.should.equal(10);
            tx.outputs.forEach((output) => output.should.deepEqual({
                coin: 'tsui:deep',
                address: testData.sender.address,
                value: amount.toString(),
            }));
            tx.suiTransaction.gasData.owner.should.equal(gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(10);
            const programmableTx = tx.suiTransaction.tx;
            // objects sent as input + no. of recipient amounts(pure arg) + no. of unique recipient addresses(de-duped objects = 1)
            programmableTx.inputs.length.should.equal(numberOfInputObjects + numberOfRecipients + 1);
            programmableTx.transactions[0].kind.should.equal('MergeCoins');
            programmableTx.transactions[0].sources.length.should.equal(numberOfInputObjects - 1);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
            const rebuilder = factory.from(rawTx);
            rebuilder.addSignature({ pub: testData.sender.publicKey }, Buffer.from(testData.sender.signatureHex));
            const rebuiltTx = await rebuilder.build();
            rebuiltTx.toBroadcastFormat().should.equal(rawTx);
            rebuiltTx.toJson().gasData.payment.length.should.equal(numberOfGasPaymentObjects);
            rebuiltTx.toJson().inputObjects.length.should.equal(numberOfInputObjects);
        });
    });
    describe('Fail', () => {
        it('should fail for invalid sender', async function () {
            const builder = factory.getTokenTransferBuilder();
            (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid or missing sender, got: randomString');
        });
        it('should fail for invalid payTx', async function () {
            const builder = factory.getTokenTransferBuilder();
            (0, should_1.default)(() => builder.send([testData.invalidRecipients[0]])).throwError('Invalid or missing address, got: randomString');
            (0, should_1.default)(() => builder.send([testData.invalidRecipients[1]])).throwError('Invalid recipient amount');
        });
        it('should fail for invalid gasData', function () {
            const builder = factory.getTokenTransferBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasOwner)).throwError(`Invalid gas address ${testData.invalidGasOwner.owner}`);
        });
        it('should fail for invalid gasBudget', function () {
            const builder = factory.getTokenTransferBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasBudget)).throwError('Invalid gas budget -1');
        });
        it('should fail for invalid gasPayment', function () {
            const builder = factory.getTokenTransferBuilder();
            const invalidGasPayment = {
                ...testData.gasDataWithoutGasPayment,
                payment: [
                    {
                        objectId: '',
                        version: -1,
                        digest: '',
                    },
                ],
            };
            (0, should_1.default)(() => builder.gasData(invalidGasPayment)).throwError('Invalid payment, invalid or missing version');
        });
        it('should fail for invalid inputObjects', function () {
            const builder = factory.getTokenTransferBuilder();
            const invalidInputObjects = [
                {
                    objectId: 'objectId',
                    version: 1,
                },
            ];
            // @ts-expect-error - testing invalid input
            (0, should_1.default)(() => builder.inputObjects(invalidInputObjects)).throwError('Invalid input object, missing digest');
        });
    });
});
