// @bitgo/sdk-coin-sui/dist/test/unit/transactionBuilder/transferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const testData = __importStar(require("../../resources/sui"));
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const transaction_1 = require("../../../src/lib/transaction");
const iface_1 = require("../../../src/lib/iface");
const constants_1 = require("../../../src/lib/constants");
describe('Sui Transfer Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsui');
    describe('Succeed', () => {
        it('should build a transfer tx', async function () {
            const txBuilder = factory.getTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.Transfer);
            txBuilder.sender(testData.sender.address);
            txBuilder.send(testData.recipients);
            txBuilder.gasData(testData.gasData);
            const tx = await txBuilder.build();
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.suiTransaction.gasData.payment.should.deepEqual(testData.coinsGasPayment);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: (testData.AMOUNT * 2).toString(),
                coin: 'tsui',
            });
            tx.outputs.length.should.equal(2);
            tx.outputs[0].should.deepEqual({
                address: testData.recipients[0].address,
                value: testData.recipients[0].amount,
                coin: 'tsui',
            });
            tx.outputs[1].should.deepEqual({
                address: testData.recipients[1].address,
                value: testData.recipients[1].amount,
                coin: 'tsui',
            });
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
            should_1.default.equal(rawTx, testData.TRANSFER);
        });
        it('should build a sponsored transfer tx with inputObjects', async function () {
            const inputObjects = [
                {
                    objectId: '0000000000000000000000001234567890abcdef1234567890abcdef12345678',
                    version: 100,
                    digest: '2B8XKQJ7mfxQPUWqJJAGjzBAzivkWKq2cEa3W8LLz1yB',
                },
                {
                    objectId: '000000000000000000000000abcdef1234567890abcdef1234567890abcdef12',
                    version: 200,
                    digest: 'DoJwXuz9oU5Y5v5vBRiTgisVTQuZQLmHZWeqJzzD5QUE',
                },
            ];
            // Create gas data with different owner (sponsor)
            const sponsoredGasData = {
                ...testData.gasData,
                owner: testData.feePayer.address, // Different from sender
            };
            const txBuilder = factory.getTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.Transfer);
            txBuilder.sender(testData.sender.address); // Sender
            txBuilder.send(testData.recipients);
            txBuilder.gasData(sponsoredGasData); // Gas paid by sponsor
            txBuilder.inputObjects(inputObjects); // Required for sponsored tx
            const tx = await txBuilder.build();
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            const suiTx = tx;
            // Verify transaction structure for sponsored transaction
            const programmableTx = suiTx.suiTransaction.tx;
            programmableTx.transactions.length.should.be.greaterThan(0);
            // Verify sponsored transaction characteristics
            suiTx.suiTransaction.sender.should.equal(testData.sender.address);
            suiTx.suiTransaction.gasData.owner.should.equal(testData.feePayer.address);
            // Should have transactions for merging/splitting and transferring when using inputObjects
            programmableTx.transactions.length.should.be.greaterThan(0);
            // Verify we have transfer operations (the exact structure varies with API versions)
            should_1.default.exist(programmableTx.transactions);
            // Verify inputObjects are not used in gas payment (they're separate)
            suiTx.suiTransaction.gasData.payment.should.not.containDeep(inputObjects);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.exist(rawTx);
            should_1.default.equal(typeof rawTx, 'string');
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
        });
        it('should build a split coin tx', async function () {
            const txBuilder = factory.getTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.Transfer);
            txBuilder.sender(testData.sender.address);
            const amount = 1000000000;
            const recipients = new Array(100).fill({ address: testData.sender.address, amount: amount.toString() });
            txBuilder.send(recipients);
            txBuilder.gasData(testData.gasData);
            const tx = await txBuilder.build();
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.suiTransaction.gasData.payment.should.deepEqual(testData.coinsGasPayment);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: (amount * 100).toString(),
                coin: 'tsui',
            });
            tx.outputs.length.should.equal(100);
            tx.outputs.forEach((output) => output.should.deepEqual({
                address: testData.sender.address,
                value: amount.toString(),
                coin: 'tsui',
            }));
        });
        it('should build a split coin tx with more than 255 input objects', async function () {
            const amount = 1000000000;
            const numberOfRecipients = 10;
            const numberOfPaymentObjects = 1000;
            const txBuilder = factory.getTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.Transfer);
            txBuilder.sender(testData.sender.address);
            const recipients = new Array(numberOfRecipients).fill({
                address: testData.sender.address,
                amount: amount.toString(),
            });
            const gasData = {
                ...testData.gasData,
                payment: testData.generateObjects(numberOfPaymentObjects),
            };
            txBuilder.send(recipients);
            txBuilder.gasData(gasData);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.Send);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: (amount * 10).toString(),
                coin: 'tsui',
            });
            tx.outputs.length.should.equal(10);
            tx.outputs.forEach((output) => output.should.deepEqual({
                coin: 'tsui',
                address: testData.sender.address,
                value: amount.toString(),
            }));
            tx.suiTransaction.gasData.owner.should.equal(gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(constants_1.MAX_GAS_OBJECTS - 1);
            const programmableTx = tx.suiTransaction.tx;
            // total objects - objects sent as gas payment + no. of recipient amounts(pure)
            // + no. of unique recipient addresses(de-duped objects)
            programmableTx.inputs.length.should.equal(numberOfPaymentObjects - (constants_1.MAX_GAS_OBJECTS - 1) + numberOfRecipients + 1);
            programmableTx.transactions[0].kind.should.equal('MergeCoins');
            programmableTx.transactions[0].sources.length.should.equal(constants_1.MAX_COMMAND_ARGS - 1);
            programmableTx.transactions[1].kind.should.equal('MergeCoins');
            programmableTx.transactions[1].sources.length.should.equal(numberOfPaymentObjects - (constants_1.MAX_COMMAND_ARGS - 1) - (constants_1.MAX_GAS_OBJECTS - 1));
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
            const rebuilder = factory.from(rawTx);
            rebuilder.addSignature({ pub: testData.sender.publicKey }, Buffer.from(testData.sender.signatureHex));
            const rebuiltTx = await rebuilder.build();
            rebuiltTx.toBroadcastFormat().should.equal(rawTx);
            rebuiltTx.toJson().gasData.payment.length.should.equal(numberOfPaymentObjects);
        });
    });
    describe('Fail', () => {
        it('should fail for invalid sender', async function () {
            const builder = factory.getTransferBuilder();
            (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid or missing sender, got: randomString');
        });
        it('should fail for invalid payTx', async function () {
            const builder = factory.getTransferBuilder();
            (0, should_1.default)(() => builder.send([testData.invalidRecipients[0]])).throwError('Invalid or missing address, got: randomString');
            (0, should_1.default)(() => builder.send([testData.invalidRecipients[1]])).throwError('Invalid recipient amount');
        });
        it('should fail for invalid gasData', function () {
            const builder = factory.getTransferBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasOwner)).throwError(`Invalid gas address ${testData.invalidGasOwner.owner}`);
        });
        it('should fail for invalid gasBudget', function () {
            const builder = factory.getTransferBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasBudget)).throwError('Invalid gas budget -1');
        });
        it('should fail for invalid gasPayment', function () {
            const builder = factory.getTransferBuilder();
            const invalidGasPayment = {
                ...testData.gasDataWithoutGasPayment,
                payment: [
                    {
                        objectId: '',
                        version: -1,
                        digest: '',
                    },
                ],
            };
            (0, should_1.default)(() => builder.gasData(invalidGasPayment)).throwError('Invalid payment, invalid or missing version');
        });
        it('should fail for invalid inputObjects', function () {
            const builder = factory.getTransferBuilder();
            const invalidInputObjects = [
                {
                    objectId: '',
                    version: -1,
                    digest: '',
                },
            ];
            (0, should_1.default)(() => builder.inputObjects(invalidInputObjects)).throwError('Invalid input object, invalid or missing version');
        });
        it('should build transfer with different gas owner but no inputObjects', async function () {
            const sponsoredGasData = {
                ...testData.gasData,
                owner: testData.feePayer.address, // Different from sender
            };
            const txBuilder = factory.getTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.Transfer);
            txBuilder.sender(testData.sender.address); // Sender
            txBuilder.send(testData.recipients);
            txBuilder.gasData(sponsoredGasData); // Gas paid by sponsor
            // Note: NOT providing inputObjects - should still work (fee sponsorship without coin sponsorship)
            const tx = await txBuilder.build();
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            const suiTx = tx;
            should_1.default.not.exist(suiTx.suiTransaction.inputObjects);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(utils_1.default.isValidRawTransaction(rawTx), true);
        });
        it('should extract inputObjects from transaction data via toJson', async function () {
            const inputObjects = [
                {
                    objectId: '0x1234567890abcdef1234567890abcdef12345678',
                    version: 100,
                    digest: '2B8XKQJ7mfxQPUWqJJAGjzBAzivkWKq2cEa3W8LLz1yB',
                },
            ];
            const sponsoredGasData = {
                ...testData.gasData,
                owner: testData.feePayer.address, // Different from sender
            };
            const txBuilder = factory.getTransferBuilder();
            txBuilder.type(iface_1.SuiTransactionType.Transfer);
            txBuilder.sender(testData.sender.address);
            txBuilder.send(testData.recipients);
            txBuilder.gasData(sponsoredGasData);
            txBuilder.inputObjects(inputObjects);
            const tx = await txBuilder.build();
            // Test that toJson extracts inputObjects from transaction structure
            const txData = tx.toJson();
            should_1.default.exist(txData.inputObjects);
            // The toJson should extract the inputObjects from the transaction structure
            // This tests our new getInputObjectsFromTx method
            should_1.default.exist(txData.inputObjects);
            Array.isArray(txData.inputObjects).should.equal(true);
        });
    });
});
