// @bitgo/sdk-coin-sui/dist/test/unit/transactionBuilder/unstakingBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const testData = __importStar(require("../../resources/sui"));
const should_1 = __importDefault(require("should"));
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const iface_1 = require("../../../src/lib/iface");
describe('Sui unstaking Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsui');
    function testUnstakingBuilder(amount) {
        describe(`Success (amount=${amount})`, () => {
            async function assertMatchesFixture(txBuilder, rebuild = true) {
                const tx = (await txBuilder.build());
                tx.suiTransaction.tx.should.eql(amount === undefined ? testData.txBlockUnstakeNoAmount : testData.txBlockUnstakeWithAmount, JSON.stringify(tx.suiTransaction.tx));
                const rawTx = tx.toBroadcastFormat();
                assert.deepStrictEqual(utils_1.default.isValidRawTransaction(rawTx), true);
                assert.deepStrictEqual(rawTx, amount === undefined ? testData.WITHDRAW_STAKED_SUI : testData.WITHDRAW_STAKED_SUI_WITH_AMOUNT);
                assert.deepStrictEqual(tx.inputs, [
                    {
                        address: '0xee6dfc3da32e21541a2aeadfcd250f8a0a23bb7abda9c8988407fc32068c3746',
                        value: amount === undefined ? 'AMOUNT_UNKNOWN' : amount.toString(),
                        coin: 'tsui',
                    },
                ]);
                assert.deepStrictEqual(tx.outputs, [
                    {
                        address: '0x9882188ba3e8070a9bb06ae9446cf607914ee8ee58ed8306a3e3afff5a1bbb71',
                        value: amount === undefined ? 'AMOUNT_UNKNOWN' : amount.toString(),
                        coin: 'tsui',
                    },
                ]);
                if (rebuild) {
                    const txBuilder = factory.from(rawTx);
                    await assertMatchesFixture(txBuilder, false);
                }
            }
            it(`should build a unstaking tx`, async function () {
                const txBuilder = factory.getUnstakingBuilder();
                txBuilder.type(iface_1.SuiTransactionType.WithdrawStake);
                txBuilder.sender(testData.sender.address);
                txBuilder.unstake({ ...testData.requestWithdrawStakedSui, amount });
                txBuilder.gasData(testData.gasData);
                await assertMatchesFixture(txBuilder);
            });
        });
    }
    testUnstakingBuilder(undefined);
    testUnstakingBuilder(1e9);
    describe('Fail', () => {
        it('should fail for invalid sender', async function () {
            const builder = factory.getUnstakingBuilder();
            (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid or missing sender, got: randomString');
        });
        it('should fail for invalid gasData', function () {
            const builder = factory.getUnstakingBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasOwner)).throwError(`Invalid gas address ${testData.invalidGasOwner.owner}`);
        });
        it('should fail for invalid gasBudget', function () {
            const builder = factory.getUnstakingBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasBudget)).throwError('Invalid gas budget -1');
        });
        it('should fail for invalid gasPayment', function () {
            const builder = factory.getUnstakingBuilder();
            const invalidGasPayment = {
                ...testData.gasDataWithoutGasPayment,
                payment: [
                    {
                        objectId: '',
                        version: -1,
                        digest: '',
                    },
                ],
            };
            (0, should_1.default)(() => builder.gasData(invalidGasPayment)).throwError('Invalid payment, invalid or missing version');
        });
    });
});
