// @bitgo/sdk-coin-sui/dist/test/unit/transactionBuilder/walrusStakingBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const testData = __importStar(require("../../resources/sui"));
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const transaction_1 = require("../../../src/lib/transaction");
const iface_1 = require("../../../src/lib/iface");
describe('Walrus Staking Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsui:wal');
    describe('Succeed', () => {
        it('should build a staking tx', async function () {
            const txBuilder = factory.getWalrusStakingBuilder();
            txBuilder.type(iface_1.SuiTransactionType.WalrusStakeWithPool);
            txBuilder.sender(testData.sender.address);
            txBuilder.stake([testData.requestWalrusStakeWithPool]);
            txBuilder.gasData(testData.gasData);
            txBuilder.inputObjects([testData.walToken]);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.StakingAdd);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: testData.STAKING_AMOUNT.toString(),
                coin: 'tsui:wal',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: testData.requestWalrusStakeWithPool.validatorAddress,
                value: testData.requestWalrusStakeWithPool.amount.toString(),
                coin: 'tsui:wal',
            });
            const rawTx = tx.toBroadcastFormat();
            utils_1.default.isValidRawTransaction(rawTx).should.be.true();
            const rebuilder = factory.from(rawTx);
            rebuilder.addSignature({ pub: testData.sender.publicKey }, Buffer.from(testData.sender.signatureHex));
            const rebuiltTx = await rebuilder.build();
            rebuiltTx.toBroadcastFormat().should.equal(rawTx);
            tx.suiTransaction.gasData.owner.should.equal(testData.gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(testData.gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(testData.gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(testData.gasData.payment.length);
            const ptb = tx.suiTransaction.tx;
            ptb.inputs.length.should.equal(5); // WAL object, Staking shared object, Amount, Validator
            ptb.transactions[0].kind.should.equal('SplitCoins'); // Only providing one WAL token
        });
        it('should build a staking tx for multiple gas objects and WAL tokens', async function () {
            const numberOfInputObjects = 100;
            const numberOfGasPaymentObjects = 10;
            const txBuilder = factory.getWalrusStakingBuilder();
            txBuilder.type(iface_1.SuiTransactionType.WalrusStakeWithPool);
            txBuilder.sender(testData.sender.address);
            txBuilder.stake([testData.requestWalrusStakeWithPool]);
            const gasData = {
                ...testData.gasData,
                payment: testData.generateObjects(numberOfGasPaymentObjects),
            };
            txBuilder.gasData(gasData);
            txBuilder.inputObjects(testData.generateObjects(numberOfInputObjects));
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.StakingAdd);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.sender.address,
                value: testData.STAKING_AMOUNT.toString(),
                coin: 'tsui:wal',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: testData.requestWalrusStakeWithPool.validatorAddress,
                value: testData.requestWalrusStakeWithPool.amount.toString(),
                coin: 'tsui:wal',
            });
            tx.suiTransaction.gasData.owner.should.equal(gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(10);
            const ptb = tx.suiTransaction.tx;
            ptb.inputs.length.should.equal(numberOfInputObjects + 4);
            ptb.transactions[0].kind.should.equal('MergeCoins'); // Merge all input objects provided
            ptb.transactions[0].sources.length.should.equal(numberOfInputObjects - 1);
            ptb.transactions[1].kind.should.equal('SplitCoins'); // Split the desired amount off of the input object
            const rawTx = tx.toBroadcastFormat();
            utils_1.default.isValidRawTransaction(rawTx).should.be.true();
            const rebuilder = factory.from(rawTx);
            rebuilder.addSignature({ pub: testData.sender.publicKey }, Buffer.from(testData.sender.signatureHex));
            const rebuiltTx = await rebuilder.build();
            rebuiltTx.toBroadcastFormat().should.equal(rawTx);
            rebuiltTx.toJson().gasData.payment.length.should.equal(numberOfGasPaymentObjects);
            rebuiltTx.toJson().inputObjects.length.should.equal(numberOfInputObjects);
        });
    });
    describe('Fail', () => {
        it('should fail for invalid sender', async function () {
            const builder = factory.getWalrusStakingBuilder();
            (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid or missing sender, got: randomString');
        });
        it('should fail for invalid gasData', function () {
            const builder = factory.getWalrusStakingBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasOwner)).throwError(`Invalid gas address ${testData.invalidGasOwner.owner}`);
        });
        it('should fail for invalid gasBudget', function () {
            const builder = factory.getWalrusStakingBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasBudget)).throwError('Invalid gas budget -1');
        });
        it('should fail for invalid gasPayment', function () {
            const builder = factory.getWalrusStakingBuilder();
            const invalidGasPayment = {
                ...testData.gasDataWithoutGasPayment,
                payment: [
                    {
                        objectId: '',
                        version: -1,
                        digest: '',
                    },
                ],
            };
            (0, should_1.default)(() => builder.gasData(invalidGasPayment)).throwError('Invalid payment, invalid or missing version');
        });
    });
});
