// @bitgo/sdk-coin-sui/dist/test/unit/transactionBuilder/walrusWithdrawBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const assert_1 = __importDefault(require("assert"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const testData = __importStar(require("../../resources/sui"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const transaction_1 = require("../../../src/lib/transaction");
const iface_1 = require("../../../src/lib/iface");
const constants_1 = require("../../../src/lib/constants");
describe('Walrus Withdraw Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsui:wal');
    describe('Succeed', () => {
        async function assertRebuild(rawTx) {
            utils_1.default.isValidRawTransaction(rawTx).should.be.true();
            const rebuilder = factory.from(rawTx);
            rebuilder.addSignature({ pub: testData.sender.publicKey }, Buffer.from(testData.sender.signatureHex));
            const rebuiltTx = await rebuilder.build();
            rebuiltTx.toBroadcastFormat().should.equal(rawTx);
        }
        it('should build a request withdraw partial tx', async function () {
            const txBuilder = factory.getWalrusRequestWithdrawStakeBuilder();
            txBuilder.type(iface_1.SuiTransactionType.WalrusRequestWithdrawStake);
            txBuilder.sender(testData.sender.address);
            txBuilder.requestWithdrawStake(testData.requestWalrusWithdrawPartial);
            txBuilder.gasData(testData.gasData);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.StakingDeactivate);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: '0xee6dfc3da32e21541a2aeadfcd250f8a0a23bb7abda9c8988407fc32068c3746',
                value: testData.STAKING_AMOUNT.toString(),
                coin: 'tsui:wal',
            });
            tx.outputs.length.should.equal(0);
            const rawTx = tx.toBroadcastFormat();
            await assertRebuild(rawTx);
            tx.suiTransaction.gasData.owner.should.equal(testData.gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(testData.gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(testData.gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(testData.gasData.payment.length);
            const ptb = tx.suiTransaction.tx;
            ptb.inputs.length.should.equal(4); // StakedWal, Amount, Sender Address, WAL Staking Shared Object
            ptb.transactions.length.should.equal(3);
            ptb.transactions[0].kind.should.equal('MoveCall');
            ptb.transactions[0].target.should.endWith('::staked_wal::split');
            ptb.transactions[1].kind.should.equal('MoveCall');
            ptb.transactions[1].target.should.endWith('::staking::request_withdraw_stake');
            ptb.transactions[2].kind.should.equal('TransferObjects');
        });
        it('should build a request withdraw full tx', async function () {
            const txBuilder = factory.getWalrusRequestWithdrawStakeBuilder();
            txBuilder.type(iface_1.SuiTransactionType.WalrusRequestWithdrawStake);
            txBuilder.sender(testData.sender.address);
            txBuilder.requestWithdrawStake(testData.requestWalrusWithdrawFull);
            txBuilder.gasData(testData.gasData);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.StakingDeactivate);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: '0xee6dfc3da32e21541a2aeadfcd250f8a0a23bb7abda9c8988407fc32068c3746',
                value: constants_1.AMOUNT_UNKNOWN_TEXT,
                coin: 'tsui:wal',
            });
            tx.outputs.length.should.equal(0);
            const rawTx = tx.toBroadcastFormat();
            await assertRebuild(rawTx);
            tx.suiTransaction.gasData.owner.should.equal(testData.gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(testData.gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(testData.gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(testData.gasData.payment.length);
            const ptb = tx.suiTransaction.tx;
            ptb.inputs.length.should.equal(2); // StakedWal, WAL Staking Shared Object
            ptb.transactions.length.should.equal(1);
            ptb.transactions[0].kind.should.equal('MoveCall');
            ptb.transactions[0].target.should.endWith('::staking::request_withdraw_stake');
        });
        it('should build a withdraw tx', async function () {
            const txBuilder = factory.getWalrusRequestWithdrawStakeBuilder();
            txBuilder.type(iface_1.SuiTransactionType.WalrusWithdrawStake);
            txBuilder.sender(testData.sender.address);
            txBuilder.requestWithdrawStake(testData.requestWalrusWithdrawFull);
            txBuilder.gasData(testData.gasData);
            const tx = await txBuilder.build();
            (0, assert_1.default)(tx instanceof transaction_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.StakingWithdraw);
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: '0xee6dfc3da32e21541a2aeadfcd250f8a0a23bb7abda9c8988407fc32068c3746',
                value: constants_1.AMOUNT_UNKNOWN_TEXT,
                coin: 'tsui:wal',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: testData.sender.address,
                value: constants_1.AMOUNT_UNKNOWN_TEXT,
                coin: 'tsui:wal',
            });
            const rawTx = tx.toBroadcastFormat();
            await assertRebuild(rawTx);
            tx.suiTransaction.gasData.owner.should.equal(testData.gasData.owner);
            tx.suiTransaction.gasData.price.should.equal(testData.gasData.price);
            tx.suiTransaction.gasData.budget.should.equal(testData.gasData.budget);
            tx.suiTransaction.gasData.payment.length.should.equal(testData.gasData.payment.length);
            const ptb = tx.suiTransaction.tx;
            ptb.inputs.length.should.equal(3); // StakedWal, Sender Address, WAL Staking Shared Object
            ptb.transactions.length.should.equal(2);
            ptb.transactions[0].kind.should.equal('MoveCall');
            ptb.transactions[0].target.should.endWith('::staking::withdraw_stake');
            ptb.transactions[1].kind.should.equal('TransferObjects');
        });
    });
    describe('Fail', () => {
        it('should fail for invalid sender', async function () {
            const builder = factory.getWalrusRequestWithdrawStakeBuilder();
            (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid or missing sender, got: randomString');
        });
        it('should fail for invalid gasData', function () {
            const builder = factory.getWalrusRequestWithdrawStakeBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasOwner)).throwError(`Invalid gas address ${testData.invalidGasOwner.owner}`);
        });
        it('should fail for invalid gasBudget', function () {
            const builder = factory.getWalrusRequestWithdrawStakeBuilder();
            (0, should_1.default)(() => builder.gasData(testData.invalidGasBudget)).throwError('Invalid gas budget -1');
        });
        it('should fail for invalid gasPayment', function () {
            const builder = factory.getWalrusRequestWithdrawStakeBuilder();
            const invalidGasPayment = {
                ...testData.gasDataWithoutGasPayment,
                payment: [
                    {
                        objectId: '',
                        version: -1,
                        digest: '',
                    },
                ],
            };
            (0, should_1.default)(() => builder.gasData(invalidGasPayment)).throwError('Invalid payment, invalid or missing version');
        });
    });
});
