// @bitgo/utxo-core/dist/cjs/src/testutil/descriptor/descriptors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultXPubs = getDefaultXPubs;
exports.getUnspendableKey = getUnspendableKey;
exports.getPsbtParams = getPsbtParams;
exports.containsKey = containsKey;
exports.getTapLeafScripts = getTapLeafScripts;
exports.getDescriptor = getDescriptor;
exports.getDescriptorMap = getDescriptorMap;
const assert_1 = __importDefault(require("assert"));
const secp256k1_1 = require("@bitgo/secp256k1");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const key_utils_1 = require("../key.utils");
function getDefaultXPubs(seed) {
    return (0, key_utils_1.getKeyTriple)(seed).map((k) => k.neutered().toBase58());
}
function getUnspendableKey() {
    /*
    https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs
  
    ```
    If one or more of the spending conditions consist of just a single key (after aggregation), the most likely one should
    be made the internal key. If no such condition exists, it may be worthwhile adding one that consists of an aggregation
    of all keys participating in all scripts combined; effectively adding an "everyone agrees" branch. If that is
    inacceptable, pick as internal key a "Nothing Up My Sleeve" (NUMS) point, i.e., a point with unknown discrete
    logarithm.
  
    One example of such a point is H = lift_x(0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0) which is
    constructed by taking the hash of the standard uncompressed encoding of the secp256k1 base point G as X coordinate.
    In order to avoid leaking the information that key path spending is not possible it is recommended to pick a fresh
    integer r in the range 0...n-1 uniformly at random and use H + rG as internal key. It is possible to prove that this
    internal key does not have a known discrete logarithm with respect to G by revealing r to a verifier who can then
    reconstruct how the internal key was created.
    ```
  
    We could do the random integer trick here, but for internal testing it is sufficient to use the fixed point.
    */
    return '50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0';
}
function toDescriptorMap(v) {
    return new Map(Object.entries(v).map(([k, v]) => [k, wasm_utxo_1.Descriptor.fromString(v, 'derivable')]));
}
function toXPub(k, path) {
    if (typeof k === 'string') {
        return k + '/' + path;
    }
    return k.neutered().toBase58() + '/' + path;
}
function toPlain(k, { xonly = false } = {}) {
    if (typeof k === 'string') {
        if (k.startsWith('xpub') || k.startsWith('xprv')) {
            return toPlain(secp256k1_1.bip32.fromBase58(k), { xonly });
        }
        return k;
    }
    return k.publicKey.subarray(xonly ? 1 : 0).toString('hex');
}
function toXOnly(k) {
    return toPlain(k, { xonly: true });
}
function multiArgs(m, n, keys, path) {
    if (n < m) {
        throw new Error(`Cannot create ${m} of ${n} multisig`);
    }
    if (keys.length < n) {
        throw new Error(`Not enough keys for ${m} of ${n} multisig: keys.length=${keys.length}`);
    }
    keys = keys.slice(0, n);
    return [m, ...keys.map((k) => toXPub(k, path))];
}
function getPsbtParams(t) {
    switch (t) {
        case 'Wsh2Of3CltvDrop':
            return { locktime: 1 };
        default:
            return {};
    }
}
function getDescriptorNode(template, keys = getDefaultXPubs(), path = '0/*') {
    switch (template) {
        case 'Wsh2Of3':
            return {
                wsh: { multi: multiArgs(2, 3, keys, path) },
            };
        case 'Wsh2Of3CltvDrop':
            const { locktime } = getPsbtParams(template);
            (0, assert_1.default)(locktime);
            return {
                wsh: {
                    and_v: [{ 'r:after': locktime }, { multi: multiArgs(2, 3, keys, path) }],
                },
            };
        case 'Wsh2Of2':
            return {
                wsh: { multi: multiArgs(2, 2, keys, path) },
            };
        case 'Tr2Of3-NoKeyPath':
            return {
                tr: [getUnspendableKey(), { multi_a: multiArgs(2, 3, keys, path) }],
            };
        case 'Tr1Of3-NoKeyPath-Tree':
            return {
                tr: [
                    getUnspendableKey(),
                    [{ pk: toXPub(keys[0], path) }, [{ pk: toXPub(keys[1], path) }, { pk: toXPub(keys[2], path) }]],
                ],
            };
        case 'Tr1Of3-NoKeyPath-Tree-Plain':
            return {
                tr: [getUnspendableKey(), [{ pk: toXOnly(keys[0]) }, [{ pk: toXOnly(keys[1]) }, { pk: toXOnly(keys[2]) }]]],
            };
    }
    throw new Error(`Unknown descriptor template: ${template}`);
}
function getTapLeafScriptNodes(t) {
    if (Array.isArray(t)) {
        if (t.length !== 2) {
            throw new Error(`expected tuple, got: ${JSON.stringify(t)}`);
        }
        return t.map((v) => (Array.isArray(v) ? getTapLeafScriptNodes(v) : v)).flat();
    }
    if (typeof t === 'object') {
        const node = t;
        if (!('tr' in node)) {
            throw new Error(`TapLeafScripts are only supported for Taproot descriptors, got: ${t}`);
        }
        if (!Array.isArray(node.tr) || node.tr.length !== 2) {
            throw new Error(`expected tuple, got: ${JSON.stringify(node.tr)}`);
        }
        const tapscript = node.tr[1];
        if (!Array.isArray(tapscript)) {
            throw new Error(`expected tapscript to be an array, got: ${JSON.stringify(tapscript)}`);
        }
        return getTapLeafScriptNodes(tapscript);
    }
    throw new Error(`Invalid input: ${JSON.stringify(t)}`);
}
function containsKey(script, key) {
    if (script instanceof wasm_utxo_1.Miniscript) {
        script = wasm_utxo_1.ast.fromMiniscript(script);
    }
    if ('pk' in script) {
        return script.pk === toXOnly(key);
    }
    throw new Error(`Unsupported script type: ${JSON.stringify(script)}`);
}
function getTapLeafScripts(d) {
    return getTapLeafScriptNodes(wasm_utxo_1.ast.fromDescriptor(d)).map((n) => wasm_utxo_1.Miniscript.fromString(wasm_utxo_1.ast.formatNode(n), 'tap').toString());
}
function getDescriptor(template, keys = getDefaultXPubs(), path = '0/*') {
    return wasm_utxo_1.Descriptor.fromStringDetectType(wasm_utxo_1.ast.formatNode(getDescriptorNode(template, keys, path)));
}
function getDescriptorMap(template, keys = getDefaultXPubs()) {
    return toDescriptorMap({
        external: getDescriptor(template, keys, '0/*').toString(),
        internal: getDescriptor(template, keys, '1/*').toString(),
    });
}
