// @bitgo/utxo-core/dist/cjs/src/testutil/descriptor/mock.utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockDerivedDescriptorWalletOutput = mockDerivedDescriptorWalletOutput;
exports.mockPsbt = mockPsbt;
exports.mockPsbtDefault = mockPsbtDefault;
exports.mockPsbtDefaultWithDescriptorTemplate = mockPsbtDefaultWithDescriptorTemplate;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const descriptor_1 = require("../../descriptor");
const descriptors_1 = require("./descriptors");
function mockOutputId(id) {
    const hash = id?.hash ?? Buffer.alloc(32, 1).toString('hex');
    const vout = id?.vout ?? 0;
    return { hash, vout };
}
function mockDerivedDescriptorWalletOutput(descriptor, outputParams = {}) {
    const { value = BigInt(1e6) } = outputParams;
    const { hash, vout } = mockOutputId(outputParams.id);
    return {
        hash,
        index: vout,
        witnessUtxo: {
            script: (0, descriptor_1.createScriptPubKeyFromDescriptor)(descriptor, undefined),
            value,
        },
        descriptor,
        selectTapLeafScript: outputParams.selectTapLeafScript,
        sequence: outputParams.sequence,
    };
}
function tryDeriveAtIndex(descriptor, index) {
    return descriptor.hasWildcard() ? descriptor.atDerivationIndex(index) : descriptor;
}
function mockPsbt(inputs, outputs, params = {}) {
    return (0, descriptor_1.createPsbt)({ ...params, network: params.network ?? utxolib.networks.bitcoin }, inputs.map((i) => mockDerivedDescriptorWalletOutput(tryDeriveAtIndex(i.descriptor, i.index), i)), outputs.map((o) => {
        const derivedDescriptor = tryDeriveAtIndex(o.descriptor, o.index);
        return {
            script: (0, descriptor_1.createScriptPubKeyFromDescriptor)(derivedDescriptor, undefined),
            value: o.value,
            descriptor: o.external ? undefined : derivedDescriptor,
        };
    }));
}
function mockPsbtDefault({ descriptorSelf = (0, descriptors_1.getDescriptor)('Wsh2Of3', (0, descriptors_1.getDefaultXPubs)('a')), descriptorOther = (0, descriptors_1.getDescriptor)('Wsh2Of3', (0, descriptors_1.getDefaultXPubs)('b')), params = {}, } = {}) {
    return mockPsbt([
        { descriptor: descriptorSelf, index: 0 },
        { descriptor: descriptorSelf, index: 1, id: { vout: 1 } },
    ], [
        { descriptor: descriptorOther, index: 0, value: BigInt(4e5), external: true },
        { descriptor: descriptorSelf, index: 0, value: BigInt(4e5) },
    ], params);
}
function mockPsbtDefaultWithDescriptorTemplate(t, params = {}) {
    return mockPsbtDefault({
        descriptorSelf: (0, descriptors_1.getDescriptor)(t, (0, descriptors_1.getDefaultXPubs)('a')),
        descriptorOther: (0, descriptors_1.getDescriptor)(t, (0, descriptors_1.getDefaultXPubs)('b')),
        params,
    });
}
