// @bitgo/sdk-core/dist/test/unit/bitgo/utils/messageTypes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../../src");
describe('Message Types', () => {
    describe('MessageStandardType', () => {
        it('should define supported message standard types', () => {
            should_1.default.exist(src_1.MessageStandardType.UNKNOWN);
            should_1.default.exist(src_1.MessageStandardType.EIP191);
            should_1.default.equal(src_1.MessageStandardType.UNKNOWN, 'UNKNOWN');
            should_1.default.equal(src_1.MessageStandardType.EIP191, 'EIP191');
        });
    });
    describe('BroadcastableMessage', () => {
        it('should validate interface structure', () => {
            const message = {
                type: src_1.MessageStandardType.EIP191,
                payload: 'Test payload',
                serializedSignatures: [
                    {
                        publicKey: 'pub1',
                        signature: 'signature1',
                    },
                    {
                        publicKey: 'pub2',
                        signature: 'signature2',
                    },
                ],
                signers: ['signer1', 'signer2'],
                metadata: { chainId: 1 },
                signablePayload: 'signable payload',
            };
            should_1.default.equal(message.type, src_1.MessageStandardType.EIP191);
            should_1.default.equal(message.payload, 'Test payload');
            should_1.default.deepEqual(message.serializedSignatures, [
                {
                    publicKey: 'pub1',
                    signature: 'signature1',
                },
                {
                    publicKey: 'pub2',
                    signature: 'signature2',
                },
            ]);
            should_1.default.deepEqual(message.signers, ['signer1', 'signer2']);
            should_1.default.deepEqual(message.metadata, { chainId: 1 });
            should_1.default.deepEqual(message.signablePayload, 'signable payload');
        });
        it('should allow optional fields to be undefined', () => {
            const message = {
                type: src_1.MessageStandardType.UNKNOWN,
                payload: 'Minimal message',
                serializedSignatures: [],
                signers: [],
            };
            should_1.default.not.exist(message.metadata);
            should_1.default.not.exist(message.signablePayload);
        });
    });
});
