// @bitgo/sdk-core/dist/test/unit/bitgo/utils/txRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("should");
const txRequest_1 = require("../../../../src/bitgo/utils/txRequest");
describe('txRequest utils', () => {
    describe('getTxRequestApiVersion', function () {
        const testCases = [
            {
                wallet: {
                    baseCoin: { getMPCAlgorithm: () => 'ecdsa' },
                    type: () => 'hot',
                    multisigType: () => 'tss',
                },
                requestedApiVersion: 'lite',
                expectedApiVersion: '',
                expectedErrorMessage: 'For ECDSA tss wallets, parameter `apiVersion` must be `full`.',
            },
            {
                wallet: {
                    baseCoin: { getMPCAlgorithm: () => 'eddsa' },
                    type: () => 'cold',
                    multisigType: () => 'tss',
                },
                requestedApiVersion: 'lite',
                expectedApiVersion: '',
                expectedErrorMessage: 'For non self-custodial (hot) tss wallets, parameter `apiVersion` must be `full`.',
            },
            {
                wallet: {
                    baseCoin: { getMPCAlgorithm: () => 'eddsa' },
                    type: () => 'hot',
                    multisigType: () => 'tss',
                },
                requestedApiVersion: undefined,
                expectedApiVersion: 'lite',
                expectedErrorMessage: '',
            },
            ...['hot', 'cold', 'custodial', 'backing'].map((walletType) => {
                return {
                    wallet: {
                        baseCoin: { getMPCAlgorithm: () => 'ecdsa' },
                        type: () => walletType,
                        multisigType: () => 'tss',
                    },
                    requestedApiVersion: 'full',
                    expectedApiVersion: 'full',
                    expectedErrorMessage: '',
                    shouldThrow: false,
                };
            }),
            ...['hot', 'cold', 'custodial', 'backing'].map((walletType) => {
                return {
                    wallet: {
                        baseCoin: { getMPCAlgorithm: () => 'ecdsa' },
                        type: () => walletType,
                        multisigType: () => 'tss',
                    },
                    requestedApiVersion: undefined,
                    expectedApiVersion: 'full',
                    expectedErrorMessage: '',
                    shouldThrow: false,
                };
            }),
            ...['hot', 'cold', 'custodial', 'backing'].map((walletType) => {
                return {
                    wallet: {
                        baseCoin: { getMPCAlgorithm: () => 'eddsa' },
                        type: () => walletType,
                        multisigType: () => 'tss',
                    },
                    requestedApiVersion: 'full',
                    expectedApiVersion: 'full',
                    expectedErrorMessage: '',
                    shouldThrow: false,
                };
            }),
            ...['cold', 'custodial', 'backing'].map((walletType) => {
                return {
                    wallet: {
                        baseCoin: { getMPCAlgorithm: () => 'eddsa' },
                        type: () => walletType,
                        multisigType: () => 'tss',
                    },
                    requestedApiVersion: undefined,
                    expectedApiVersion: 'full',
                    expectedErrorMessage: '',
                    shouldThrow: false,
                };
            }),
        ];
        testCases.forEach((testCase) => {
            if (testCase.expectedErrorMessage) {
                it(`should throw an error if requested apiVersion is ${testCase.requestedApiVersion} for wallet type ${testCase.wallet.type()} for a ${testCase.wallet.baseCoin.getMPCAlgorithm()} wallet`, () => {
                    (() => (0, txRequest_1.getTxRequestApiVersion)(testCase.wallet, testCase.requestedApiVersion)).should.throw(testCase.expectedErrorMessage);
                });
            }
            else {
                it(`should return ${testCase.expectedApiVersion} if requested apiVersion is ${testCase.requestedApiVersion} for wallet type ${testCase.wallet.type()} for a ${testCase.wallet.baseCoin.getMPCAlgorithm()} wallet`, () => {
                    (0, txRequest_1.getTxRequestApiVersion)(testCase.wallet, testCase.requestedApiVersion).should.equal(testCase.expectedApiVersion);
                });
            }
        });
    });
});
