// @bitgo/public-types/dist/src/schema/mpcv2/smc/keygenState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyCreationMPCv2State = exports.KeyCreationMPCv2StateEnum = void 0;
const t = __importStar(require("io-ts"));
var KeyCreationMPCv2StateEnum;
(function (KeyCreationMPCv2StateEnum) {
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC1Round1Data"] = -1] = "WaitingForOVC1Round1Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC2Round1Data"] = 0] = "WaitingForOVC2Round1Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForBitgoRound1Data"] = 1] = "WaitingForBitgoRound1Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC1Round2Data"] = 2] = "WaitingForOVC1Round2Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC2Round2Data"] = 3] = "WaitingForOVC2Round2Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForBitgoRound2Data"] = 4] = "WaitingForBitgoRound2Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC1Round3aData"] = 5] = "WaitingForOVC1Round3aData";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC2Round3Data"] = 6] = "WaitingForOVC2Round3Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC1Round3bData"] = 7] = "WaitingForOVC1Round3bData";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForBitgoRound3Data"] = 8] = "WaitingForBitgoRound3Data";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC1GenerateKey"] = 9] = "WaitingForOVC1GenerateKey";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["WaitingForOVC2GenerateKey"] = 10] = "WaitingForOVC2GenerateKey";
    KeyCreationMPCv2StateEnum[KeyCreationMPCv2StateEnum["KeyGenerationComplete"] = 11] = "KeyGenerationComplete";
})(KeyCreationMPCv2StateEnum || (exports.KeyCreationMPCv2StateEnum = KeyCreationMPCv2StateEnum = {}));
exports.KeyCreationMPCv2State = t.union([
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC1Round1Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC2Round1Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForBitgoRound1Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC1Round2Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC2Round2Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForBitgoRound2Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC1Round3aData),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC2Round3Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC1Round3bData),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForBitgoRound3Data),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC1GenerateKey),
    t.literal(KeyCreationMPCv2StateEnum.WaitingForOVC2GenerateKey),
    t.literal(KeyCreationMPCv2StateEnum.KeyGenerationComplete),
]);
