// @bitgo/abstract-utxo/dist/cjs/src/offlineVault/OfflineVaultSignable.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfflineVaultSignable = exports.XPubWithDerivationPath = void 0;
exports.toKeyTriple = toKeyTriple;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const t = __importStar(require("io-ts"));
exports.XPubWithDerivationPath = t.intersection([t.type({ xpub: t.string }), t.partial({ derivedFromParentWithSeed: t.string })], 'XPubWithDerivationPath');
/**
 * This is the transaction payload that is sent to the offline vault to sign.
 */
exports.OfflineVaultSignable = t.type({
    xpubsWithDerivationPath: t.type({
        user: exports.XPubWithDerivationPath,
        backup: exports.XPubWithDerivationPath,
        bitgo: exports.XPubWithDerivationPath,
    }),
    coinSpecific: t.type({ txHex: t.string }),
}, 'BaseTransaction');
function toKeyTriple(xpubs) {
    return [wasm_utxo_1.BIP32.fromBase58(xpubs.user.xpub), wasm_utxo_1.BIP32.fromBase58(xpubs.backup.xpub), wasm_utxo_1.BIP32.fromBase58(xpubs.bitgo.xpub)];
}
