// bitgo/dist/test/v2/resources/amsTokenConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reducedAmsTokenConfig = void 0;
exports.reducedAmsTokenConfig = {
    'hteth:faketoken': [
        {
            id: '49ff49ea-3355-4717-bbb0-5e8f5cae2201',
            fullName: 'Hoodi Testnet fake token',
            name: 'hteth:faketoken',
            prefix: '',
            suffix: 'HTETH:FAKETOKEN',
            baseUnit: 'wei',
            kind: 'crypto',
            family: 'eth',
            isToken: true,
            additionalFeatures: ['bulk-transaction'],
            excludedFeatures: ['custody-bitgo-korea'],
            decimalPlaces: 6,
            asset: 'hteth:faketoken',
            network: {
                name: 'Hoodi',
            },
            primaryKeyCurve: 'secp256k1',
            contractAddress: '0x89a959b9184b4f8c8633646d5dfd049d2ebc983a',
        },
    ],
    'tip:faketoken': [
        {
            id: 'a1b2c3d4-e5f6-4789-8abc-def123456789',
            fullName: 'Story Testnet Faketoken',
            name: 'tip:faketoken',
            prefix: '',
            suffix: 'TIP:FAKETOKEN',
            baseUnit: 'wei',
            kind: 'crypto',
            family: 'ip',
            isToken: true,
            additionalFeatures: [],
            excludedFeatures: [],
            decimalPlaces: 18,
            asset: 'tip:faketoken',
            network: {
                name: 'BaseChainTestnet',
            },
            primaryKeyCurve: 'secp256k1',
            contractAddress: '0x1234567890123456789012345678901234567890',
        },
    ],
    'thypeevm:faketoken': [
        {
            id: 'b2c3d4e5-f6a7-4890-9bcd-ef012345678a',
            fullName: 'Hyperliquid EVM Testnet Faketoken',
            name: 'thypeevm:faketoken',
            prefix: '',
            suffix: 'THYPEEVM:FAKETOKEN',
            baseUnit: 'wei',
            kind: 'crypto',
            family: 'hypeevm',
            isToken: true,
            additionalFeatures: [],
            excludedFeatures: [],
            decimalPlaces: 18,
            asset: 'thypeevm:faketoken',
            network: {
                name: 'HyperliquidTestnet',
            },
            primaryKeyCurve: 'secp256k1',
            contractAddress: '0xabcdefabcdefabcdefabcdefabcdefabcdefabcd',
        },
    ],
};
