// @bitgo/sdk-coin-tempo/dist/src/tempo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tempo = void 0;
/**
 * @prettier
 */
const abstract_eth_1 = require("@bitgo/abstract-eth");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
const utils_1 = require("./lib/utils");
const url = __importStar(require("url"));
const querystring = __importStar(require("querystring"));
class Tempo extends abstract_eth_1.AbstractEthLikeNewCoins {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
    }
    /**
     * Factory method to create Tempo instance
     */
    static createInstance(bitgo, staticsCoin) {
        return new Tempo(bitgo, staticsCoin);
    }
    /**
     * Get the chain identifier
     */
    getChain() {
        return this._staticsCoin?.name || 'tempo';
    }
    /**
     * Get the full chain name
     */
    getFullName() {
        return 'Tempo';
    }
    /** @inheritdoc */
    getBaseFactor() {
        return 1e6;
    }
    /**
     * Check if value-less transfers are allowed
     * TODO: Update based on Tempo requirements
     */
    valuelessTransferAllowed() {
        return false;
    }
    /**
     * Check if TSS is supported
     */
    supportsTss() {
        return true;
    }
    /**
     * Get the MPC algorithm (ECDSA for EVM chains)
     */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /**
     * Evaluates whether an address string is valid for Tempo
     * Supports addresses with optional memoId query parameter (e.g., 0x...?memoId=123)
     * @param address - The address to validate
     * @returns true if address is valid
     */
    isValidAddress(address) {
        if (typeof address !== 'string') {
            return false;
        }
        try {
            const { baseAddress } = this.getAddressDetails(address);
            return abstract_eth_1.optionalDeps.ethUtil.isValidAddress(abstract_eth_1.optionalDeps.ethUtil.addHexPrefix(baseAddress));
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Parse address into base address and optional memoId
     * Throws InvalidAddressError for invalid address formats
     * @param address - Address string, potentially with ?memoId=X suffix
     * @returns Object containing address, baseAddress, and memoId (null if not present)
     * @throws InvalidAddressError if address format is invalid
     */
    getAddressDetails(address) {
        if (typeof address !== 'string') {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        const destinationDetails = url.parse(address);
        const baseAddress = destinationDetails.pathname || '';
        // No query string - just a plain address
        if (destinationDetails.pathname === address) {
            return {
                address,
                baseAddress: address,
                memoId: null,
            };
        }
        // Has query string - must contain memoId
        if (!destinationDetails.query) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        const queryDetails = querystring.parse(destinationDetails.query);
        // Query string must contain memoId
        if (!queryDetails.memoId) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}, unknown query parameters`);
        }
        // Only one memoId allowed
        if (Array.isArray(queryDetails.memoId)) {
            throw new sdk_core_1.InvalidAddressError(`memoId may only be given at most once, but found ${queryDetails.memoId.length} instances in address ${address}`);
        }
        // Reject if there are other query parameters besides memoId
        const queryKeys = Object.keys(queryDetails);
        if (queryKeys.length !== 1) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}, only memoId query parameter is allowed`);
        }
        // Validate memoId format
        if (!this.isValidMemoId(queryDetails.memoId)) {
            throw new sdk_core_1.InvalidMemoIdError(`invalid address: '${address}', memoId is not valid`);
        }
        return {
            address,
            baseAddress,
            memoId: queryDetails.memoId,
        };
    }
    /**
     * Validate that a memoId is a valid non-negative integer string
     * @param memoId - The memoId to validate
     * @returns true if valid
     */
    isValidMemoId(memoId) {
        return (0, utils_1.isValidMemoId)(memoId);
    }
    /**
     * Tempo uses memoId-based addresses rather than forwarder contracts.
     * Verify that the address belongs to this wallet by checking that the
     * base (EVM) portion matches the wallet's base address.
     */
    async isWalletAddress(params) {
        const { address, baseAddress } = params;
        const rootAddress = params.rootAddress;
        if (!address) {
            throw new sdk_core_1.InvalidAddressError('address is required');
        }
        if (!this.isValidAddress(address)) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        const { baseAddress: addressBase } = this.getAddressDetails(address);
        const walletBaseAddress = baseAddress || rootAddress;
        if (!walletBaseAddress) {
            throw new sdk_core_1.InvalidAddressError('baseAddress or rootAddress is required for verification');
        }
        const { baseAddress: walletBase } = this.getAddressDetails(walletBaseAddress);
        if (addressBase.toLowerCase() !== walletBase.toLowerCase()) {
            throw new sdk_core_1.UnexpectedAddressError(`address validation failure: expected ${walletBase} but got ${addressBase}`);
        }
        return true;
    }
    /**
     * Parse a serialised Tempo transaction and return its operations as SDK outputs.
     * @inheritdoc
     */
    async parseTransaction(params) {
        const txHex = (params.txHex || params.halfSigned?.txHex);
        if (!txHex) {
            return {};
        }
        if (!(0, utils_1.isTip20Transaction)(txHex)) {
            return {};
        }
        const txBuilder = this.getTransactionBuilder();
        txBuilder.from(txHex);
        const tx = (await txBuilder.build());
        return {
            inputs: tx.inputs.map((input) => ({
                address: input.address,
                amount: input.value,
                coin: this.getChain(),
            })),
            outputs: tx.outputs.map((output) => ({
                address: output.address,
                amount: output.value,
                coin: this.getChain(),
            })),
        };
    }
    /**
     * Verify that a Tempo transaction matches the intended recipients and amounts.
     * @inheritdoc
     */
    async verifyTransaction(params) {
        const { txParams, txPrebuild } = params;
        if (!txPrebuild?.txHex) {
            return true;
        }
        // signableHex may arrive without the 0x prefix (e.g. from the TSS signing flow).
        const txHex = txPrebuild.txHex.startsWith('0x') ? txPrebuild.txHex : '0x' + txPrebuild.txHex;
        if (!(0, utils_1.isTip20Transaction)(txHex)) {
            throw new Error(`Invalid Tempo transaction hex: expected a 0x76 TIP-20 transaction`);
        }
        const txBuilder = this.getTransactionBuilder();
        txBuilder.from(txHex);
        const tx = (await txBuilder.build());
        const operations = tx.getOperations();
        // If the caller specified explicit recipients, verify they match the operations 1-to-1
        const recipients = txParams?.recipients;
        if (recipients && recipients.length > 0) {
            if (operations.length !== recipients.length) {
                throw new Error(`Transaction has ${operations.length} operation(s) but ${recipients.length} recipient(s) were requested`);
            }
            for (let i = 0; i < operations.length; i++) {
                const op = operations[i];
                const recipient = recipients[i];
                if (op.to.toLowerCase() !== recipient.address.toLowerCase()) {
                    throw new Error(`Operation ${i} recipient mismatch: expected ${recipient.address}, got ${op.to}`);
                }
                // Compare amounts in base units (smallest denomination)
                const opAmountBaseUnits = (0, utils_1.amountToTip20Units)(op.amount).toString();
                if (opAmountBaseUnits !== recipient.amount.toString()) {
                    throw new Error(`Operation ${i} amount mismatch: expected ${recipient.amount}, got ${opAmountBaseUnits}`);
                }
            }
        }
        return true;
    }
    /**
     * Build unsigned sweep transaction for TSS
     * TODO: Implement sweep transaction logic
     */
    async buildUnsignedSweepTxnTSS(params) {
        // TODO: Implement when recovery logic is needed
        // Return dummy value to prevent downstream services from breaking
        return {};
    }
    /**
     * Query block explorer for recovery information
     * TODO: Implement when Tempo block explorer is available
     */
    async recoveryBlockchainExplorerQuery(query, apiKey) {
        // TODO: Implement with Tempo block explorer API
        // Return empty object to prevent downstream services from breaking
        return {};
    }
    /**
     * Get transaction builder for Tempo
     * Returns a TIP-20 transaction builder for Tempo-specific operations
     * @param common - Optional common chain configuration
     * @protected
     */
    getTransactionBuilder(common) {
        return new lib_1.Tip20TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
}
exports.Tempo = Tempo;
