// @bitgo/sdk-coin-tempo/dist/src/tip20Token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tip20Token = void 0;
const statics_1 = require("@bitgo/statics");
const tempo_1 = require("./tempo");
const utils_1 = require("./lib/utils");
/**
 * TIP20 Token Implementation (Skeleton)
 *
 * This is a minimal skeleton for TIP20 tokens on Tempo blockchain.
 *
 * TODO: All methods will be implemented progressively
 */
class Tip20Token extends tempo_1.Tempo {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('tempo') : statics_1.coins.get('ttempo');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
        this.contractAddress = tokenConfig.tokenContractAddress;
    }
    /**
     * Create a coin constructor for a specific token
     */
    static createTokenConstructor(config) {
        return (bitgo) => new Tip20Token(bitgo, config);
    }
    /**
     * Create token constructors for all TIP20 tokens
     * @param tokenConfigs - Array of token configurations; defaults to all TIP20 tokens from statics
     */
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.tempo.tokens, ...statics_1.tokens.testnet.tempo.tokens]) {
        const configs = tokenConfigs;
        const tokensCtors = [];
        for (const token of configs) {
            const tokenConstructor = Tip20Token.createTokenConstructor(token);
            // Register by token type
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
            // Also register by contract address for lookups
            tokensCtors.push({ name: token.tokenContractAddress, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    /** Get the token type */
    get type() {
        return this.tokenConfig.type;
    }
    /** Get the token name */
    get name() {
        return this.tokenConfig.name;
    }
    /** Get the base coin */
    get coin() {
        return this.tokenConfig.coin;
    }
    /** Get the network */
    get network() {
        return this.tokenConfig.network;
    }
    /** Get the token contract address */
    get tokenContractAddress() {
        return this.tokenConfig.tokenContractAddress;
    }
    /** Get token decimal places */
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    /** @inheritDoc */
    getChain() {
        return this.tokenConfig.type;
    }
    /** @inheritDoc */
    getBaseChain() {
        return this.coin;
    }
    /** @inheritDoc */
    getFullName() {
        return 'TIP20 Token';
    }
    /** @inheritDoc */
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    /** @inheritDoc */
    valuelessTransferAllowed() {
        return false;
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /**
     * Verify that the transaction coin/token matches this token
     * @param txPrebuild - Transaction prebuild object
     * @returns true if valid, false otherwise
     */
    verifyCoin(txPrebuild) {
        if (!txPrebuild) {
            return false;
        }
        // Check if the coin or token matches this token's configuration
        const coinMatch = txPrebuild.coin === this.tokenConfig.coin || txPrebuild.coin === this.tokenConfig.type;
        const tokenMatch = !txPrebuild.token ||
            txPrebuild.token === this.tokenConfig.tokenContractAddress ||
            txPrebuild.token === this.tokenConfig.type;
        return coinMatch && tokenMatch;
    }
    /**
     * Get send method arguments for TIP-20 token transfer with memo
     * @param txInfo - Transaction information including recipient and amount
     * @returns Array of send method arguments for ABI encoding
     */
    getSendMethodArgs(txInfo) {
        const { recipient } = txInfo;
        if (!recipient) {
            throw new Error('Recipient is required for token transfer');
        }
        if (!(0, utils_1.isValidAddress)(recipient.address)) {
            throw new Error(`Invalid recipient address: ${recipient.address}`);
        }
        if (!(0, utils_1.isValidTip20Amount)(recipient.amount)) {
            throw new Error(`Invalid amount: ${recipient.amount}`);
        }
        const memo = recipient.memo;
        if (memo && Buffer.byteLength(memo, 'utf-8') > 32) {
            throw new Error('Memo too long: maximum 32 bytes');
        }
        const amountInUnits = (0, utils_1.amountToTip20Units)(recipient.amount);
        const data = (0, utils_1.encodeTip20TransferWithMemo)(recipient.address, amountInUnits, memo);
        return [
            {
                name: 'toAddress',
                type: 'address',
                value: recipient.address,
            },
            {
                name: 'value',
                type: 'uint',
                value: recipient.amount,
            },
            {
                name: 'tokenContractAddress',
                type: 'address',
                value: this.tokenConfig.tokenContractAddress,
            },
            {
                name: 'data',
                type: 'bytes',
                value: Buffer.from(data.slice(2), 'hex'),
            },
            {
                name: 'expireTime',
                type: 'uint',
                value: txInfo.expireTime,
            },
            {
                name: 'sequenceId',
                type: 'uint',
                value: txInfo.contractSequenceId,
            },
            {
                name: 'signature',
                type: 'bytes',
                value: Buffer.from(txInfo.signature.replace('0x', ''), 'hex'),
            },
        ];
    }
    /**
     * Get operation object for TIP-20 token transfer (for batch transactions)
     * @param recipient - Recipient information with address, amount, and optional memo
     * @param expireTime - Transaction expiration time
     * @param contractSequenceId - Contract sequence ID
     * @returns Operation array for ABI encoding
     */
    getOperation(recipient, expireTime, contractSequenceId) {
        if (!(0, utils_1.isValidAddress)(recipient.address)) {
            throw new Error(`Invalid recipient address: ${recipient.address}`);
        }
        if (!(0, utils_1.isValidTip20Amount)(recipient.amount)) {
            throw new Error(`Invalid amount: ${recipient.amount}`);
        }
        // Validate memo byte length (handles multi-byte UTF-8 characters)
        if (recipient.memo !== undefined && recipient.memo !== null && recipient.memo !== '') {
            const memoByteLength = new TextEncoder().encode(recipient.memo).length;
            if (memoByteLength > 32) {
                throw new Error(`Memo too long: ${memoByteLength} bytes. Maximum 32 bytes.`);
            }
        }
        const amountInUnits = (0, utils_1.amountToTip20Units)(recipient.amount);
        const data = (0, utils_1.encodeTip20TransferWithMemo)(recipient.address, amountInUnits, recipient.memo);
        // Return format compatible with parent class for ABI encoding
        return [
            ['address', 'bytes'],
            [this.tokenConfig.tokenContractAddress, Buffer.from(data.slice(2), 'hex')],
        ];
    }
}
exports.Tip20Token = Tip20Token;
