// @bitgo/sdk-coin-algo/dist/test/unit/lib/transactionBuilder/base.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = __importDefault(require("crypto"));
const statics_1 = require("@bitgo/statics");
const algosdk_1 = __importDefault(require("algosdk"));
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const sinon_1 = __importStar(require("sinon"));
const lib_1 = require("../../../../src/lib");
const AlgoResources = __importStar(require("../../../fixtures/resources"));
const STANDARD_REQUIRED_NUMBER_OF_SIGNERS = 2;
class StubTransactionBuilder extends lib_1.TransactionBuilder {
    constructor(coinConfig) {
        super(coinConfig);
    }
    getFee() {
        return this._fee;
    }
    getSender() {
        return this._sender;
    }
    getGenesisHash() {
        return this._genesisHash;
    }
    getGenesisId() {
        return this._genesisId;
    }
    getFirstRound() {
        return this._firstRound;
    }
    getLastRound() {
        return this._lastRound;
    }
    getLease() {
        return this._lease;
    }
    getNote() {
        return this._note;
    }
    getReKeyTo() {
        return this._reKeyTo;
    }
    getKeyPairs() {
        return this._keyPairs;
    }
    getTransaction() {
        return this._transaction;
    }
    buildImplementation() {
        return super.buildImplementation();
    }
    fromImplementation(rawTransaction) {
        return super.fromImplementation(rawTransaction);
    }
    signImplementation(key) {
        return super.signImplementation(key);
    }
    getSuggestedParams() {
        return this.suggestedParams;
    }
    buildAlgoTxn() {
        throw new Error('Method not implemented.');
    }
    get transactionType() {
        throw new Error('Method not implemented.');
    }
}
describe('Algo Transaction Builder', () => {
    let txnBuilder;
    const { accounts: { account1, account2, account3 }, networks: { testnet }, } = AlgoResources;
    beforeEach(() => {
        const config = statics_1.coins.get('algo');
        txnBuilder = new StubTransactionBuilder(config);
    });
    describe('setter validation', () => {
        it('should validate fee is not lt 1000 microalgos if flat fee is set to true', () => {
            txnBuilder.isFlatFee(true);
            assert_1.default.throws(() => txnBuilder.fee({ fee: '999' }), (e) => e.name === lib_1.InsufficientFeeError.name);
            should_1.default.doesNotThrow(() => txnBuilder.fee({ fee: '1000' }));
        });
        it('should validate sender address is a valid algo address', () => {
            const spy = sinon_1.default.spy(txnBuilder, 'validateAddress');
            assert_1.default.throws(() => txnBuilder.sender({ address: 'asdf' }), (e) => e.name === lib_1.AddressValidationError.name);
            should_1.default.doesNotThrow(() => txnBuilder.sender({ address: account1.address }));
            sinon_1.assert.calledTwice(spy);
        });
        it('should validate number of signers is not less than 0', () => {
            assert_1.default.throws(() => txnBuilder.numberOfRequiredSigners(-1), /Number of signers: '-1' cannot be negative/);
            for (let i = 0; i < STANDARD_REQUIRED_NUMBER_OF_SIGNERS; i++) {
                should_1.default.doesNotThrow(() => txnBuilder.numberOfRequiredSigners(i));
            }
        });
    });
    describe('suggested params verification', () => {
        it('should retrieve the suggested parameters as they have been set', () => {
            const isFlatFee = true;
            const fee = 1000;
            const firstRound = 1;
            const lastRound = 10;
            const genesisId = 'testnet-v1.0';
            const genesisHash = 'SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=';
            txnBuilder
                .isFlatFee(isFlatFee)
                .fee({ fee: fee.toString() })
                .firstRound(firstRound)
                .lastRound(lastRound)
                .genesisId(genesisId)
                .genesisHash(genesisHash);
            const suggestedParams = txnBuilder.getSuggestedParams();
            should_1.default.equal(isFlatFee, suggestedParams.flatFee);
            should_1.default.equal(fee, suggestedParams.fee);
            should_1.default.equal(firstRound, suggestedParams.firstRound);
            should_1.default.equal(lastRound, suggestedParams.lastRound);
            should_1.default.equal(genesisId, suggestedParams.genesisID);
            should_1.default.equal(genesisHash, suggestedParams.genesisHash);
        });
    });
    describe('private key validation', () => {
        it('validates byte arrays', () => {
            should_1.default.doesNotThrow(() => txnBuilder.validateKey({ key: account1.secretKey }));
        });
        it('validates hex encoded strings', () => {
            should_1.default.doesNotThrow(() => txnBuilder.validateKey({ key: account1.secretKey.toString('hex') }));
        });
        it('validates base64 encoded strings', () => {
            should_1.default.doesNotThrow(() => txnBuilder.validateKey({ key: account1.secretKey.toString('base64') }));
        });
    });
    describe('implementation functions', () => {
        const to = account1.address;
        const from = account2.address;
        const reKeyTo = account3.address;
        const amount = 1000;
        const firstRound = 1;
        const lastRound = 10;
        const closeRemainderTo = account3.address;
        // Uint8array conversion required because algosdk checks if the constructor
        // is Uint8Array.
        const lease = new Uint8Array(crypto_1.default.randomBytes(32));
        const note = new Uint8Array(Buffer.from('note', 'utf-8'));
        const fee = 1000;
        const algoTxn = algosdk_1.default.makePaymentTxnWithSuggestedParams(from, to, amount, closeRemainderTo, note, {
            fee,
            flatFee: true,
            firstRound,
            lastRound,
            genesisID: testnet.genesisID,
            genesisHash: testnet.genesisHash,
        }, reKeyTo);
        algoTxn.fee = fee;
        algoTxn.flatFee = true;
        algoTxn.addLease(lease);
        it('should assign all decoded fields into transaction builder', () => {
            txnBuilder.fromImplementation(algosdk_1.default.encodeUnsignedTransaction(algoTxn));
            (0, should_1.default)(txnBuilder.getFee()).equal(fee);
            (0, should_1.default)(txnBuilder.getSender()).equal(from);
            (0, should_1.default)(txnBuilder.getGenesisHash()).equal(testnet.genesisHash);
            (0, should_1.default)(txnBuilder.getGenesisId()).equal(testnet.genesisID);
            (0, should_1.default)(txnBuilder.getFirstRound()).equal(firstRound);
            (0, should_1.default)(txnBuilder.getLastRound()).equal(lastRound);
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            (0, should_1.default)(Buffer.from(txnBuilder.getLease()).toString('hex')).equal(Buffer.from(lease).toString('hex'));
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            (0, should_1.default)(Buffer.from(txnBuilder.getNote()).toString('hex')).equal(Buffer.from(note).toString('hex'));
            (0, should_1.default)(txnBuilder.getReKeyTo()).equal(reKeyTo);
            (0, should_1.default)(txnBuilder.getTransaction().getAlgoTransaction()).not.be.undefined();
        });
        it('should assign signers into transaction builder', () => {
            sinon_1.default.stub(lib_1.algoUtils, 'decodeAlgoTxn').callsFake((txnBytes) => {
                return {
                    rawTransaction: Buffer.from([]),
                    signed: true,
                    signers: [
                        '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5QTE',
                        '7RI43DTCDCQ2HDNEP3IAEMHQLAEBN3ITXIZQHLC55OIRKSEQQAS52OYKJE',
                    ],
                    txn: algoTxn,
                };
            });
            txnBuilder.fromImplementation(algosdk_1.default.encodeUnsignedTransaction(algoTxn));
            (0, should_1.default)(txnBuilder.getTransaction().signers).deepEqual([
                '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5QTE',
                '7RI43DTCDCQ2HDNEP3IAEMHQLAEBN3ITXIZQHLC55OIRKSEQQAS52OYKJE',
            ]);
            sinon_1.default.restore();
        });
        it('transaction builder should fail when signers are incorrect', () => {
            sinon_1.default.stub(lib_1.algoUtils, 'decodeAlgoTxn').callsFake((txnBytes) => {
                return {
                    rawTransaction: Buffer.from([]),
                    signed: true,
                    signers: [
                        '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5AAA',
                        '7RI43DTCDCQ2HDNEP3IAEMHQLAEBN3ITXIZQHLC55OIRKSEQQAS52OYKJE',
                    ],
                    txn: algoTxn,
                };
            });
            (0, should_1.default)(() => txnBuilder.fromImplementation(algosdk_1.default.encodeUnsignedTransaction(algoTxn))).throw("The address '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5AAA' is not a well-formed algorand address");
            sinon_1.default.restore();
        });
        it('should not sign the transaction', () => {
            const txn = txnBuilder.getTransaction();
            txn.setAlgoTransaction(algoTxn);
            txn.setNumberOfRequiredSigners(1);
            assert_1.default.throws(() => txnBuilder.signImplementation({ key: Buffer.from(account1.prvKey).toString('hex') }), new RegExp('Invalid base32 characters'));
        });
        it('should sign the transaction', () => {
            const { prv, pub } = new lib_1.KeyPair().recordKeysFromPrivateKeyInProtocolFormat(account1.prvKey);
            const prvKey = lib_1.algoUtils.encodeSeed(Buffer.from(prv + pub));
            const txn = txnBuilder.getTransaction();
            txn.setAlgoTransaction(algoTxn);
            txn.setNumberOfRequiredSigners(1);
            should_1.default.doesNotThrow(() => txnBuilder.signImplementation({ key: prvKey }));
        });
    });
    describe('transaction validation', () => {
        it('should validate a normal transaction', () => {
            txnBuilder
                .fee({ fee: '1000' })
                .isFlatFee(true)
                .firstRound(1)
                .lastRound(10)
                .sender({ address: account1.address })
                .genesisId(testnet.genesisID)
                .genesisHash(testnet.genesisHash);
            should_1.default.doesNotThrow(() => txnBuilder.validateTransaction(txnBuilder.getTransaction()));
        });
        it('should validate last round is after first round', () => {
            txnBuilder
                .fee({ fee: '1000' })
                .isFlatFee(true)
                .firstRound(10)
                .lastRound(1)
                .sender({ address: account1.address })
                .genesisId(testnet.genesisID)
                .genesisHash(testnet.genesisHash);
            assert_1.default.throws(() => txnBuilder.validateTransaction(txnBuilder.getTransaction()), new RegExp('Transaction validation failed: "value" failed custom validation because lastRound cannot be greater than or equal to firstRound'));
        });
        it('should build a normal transaction with correct signers', () => {
            txnBuilder
                .fee({ fee: '1000' })
                .isFlatFee(true)
                .firstRound(1)
                .lastRound(10)
                .sender({ address: account1.address })
                .setSigners([
                '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5QTE',
                '7RI43DTCDCQ2HDNEP3IAEMHQLAEBN3ITXIZQHLC55OIRKSEQQAS52OYKJE',
            ])
                .genesisId(testnet.genesisID)
                .genesisHash(testnet.genesisHash);
            (0, should_1.default)(txnBuilder.getTransaction().signers).deepEqual([
                '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5QTE',
                '7RI43DTCDCQ2HDNEP3IAEMHQLAEBN3ITXIZQHLC55OIRKSEQQAS52OYKJE',
            ]);
        });
        it('should not build a normal transaction with incorrect signers', () => {
            (0, should_1.default)(() => txnBuilder
                .fee({ fee: '1000' })
                .isFlatFee(true)
                .firstRound(1)
                .lastRound(10)
                .sender({ address: account1.address })
                .setSigners([
                '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5AAA',
                '7RI43DTCDCQ2HDNEP3IAEMHQLAEBN3ITXIZQHLC55OIRKSEQQAS52OYKJE',
            ])
                .genesisId(testnet.genesisID)
                .genesisHash(testnet.genesisHash)).throw("The address '25NJQAMCWEFLPVKL73J4SZAHHIHOC4XT3KTCGJNPAINGR5YHKENMEF5AAA' is not a well-formed algorand address");
        });
    });
});
