// @bitgo/public-types/src/schema/mpcv2/keyGen/keygenRound.ts
import * as t from "io-ts";
import { NonEmptyString } from "io-ts-types";
import { MPCv2BroadcastMessage, MPCv2P2PMessage } from "../common";

export const MPCv2KeyGenRound1Request = t.type({
  userGpgPublicKey: NonEmptyString,
  backupGpgPublicKey: NonEmptyString,
  userMsg1: MPCv2BroadcastMessage,
  backupMsg1: MPCv2BroadcastMessage,
});

export type MPCv2KeyGenRound1Request = t.TypeOf<
  typeof MPCv2KeyGenRound1Request
>;

export const MPCv2KeyGenRound1Response = t.type({
  walletGpgPubKeySigs: NonEmptyString,
  sessionId: NonEmptyString,
  bitgoMsg1: MPCv2BroadcastMessage,
  bitgoToUserMsg2: MPCv2P2PMessage,
  bitgoToBackupMsg2: MPCv2P2PMessage,
});

export type MPCv2KeyGenRound1Response = t.TypeOf<
  typeof MPCv2KeyGenRound1Response
>;

export const MPCv2KeyGenRound2Request = t.type({
  sessionId: NonEmptyString,
  userMsg2: MPCv2P2PMessage,
  userCommitment2: NonEmptyString,
  backupMsg2: MPCv2P2PMessage,
  backupCommitment2: NonEmptyString,
});

export type MPCv2KeyGenRound2Request = t.TypeOf<
  typeof MPCv2KeyGenRound2Request
>;

export const MPCv2KeyGenRound2Response = t.type({
  sessionId: NonEmptyString,
  bitgoCommitment2: NonEmptyString,
  bitgoToUserMsg3: MPCv2P2PMessage,
  bitgoToBackupMsg3: MPCv2P2PMessage,
});

export type MPCv2KeyGenRound2Response = t.TypeOf<
  typeof MPCv2KeyGenRound2Response
>;

export const MPCv2KeyGenRound3Request = t.type({
  sessionId: NonEmptyString,
  userMsg3: MPCv2P2PMessage,
  backupMsg3: MPCv2P2PMessage,
  userMsg4: MPCv2BroadcastMessage,
  backupMsg4: MPCv2BroadcastMessage,
});

export type MPCv2KeyGenRound3Request = t.TypeOf<
  typeof MPCv2KeyGenRound3Request
>;

export const MPCv2KeyGenRound3Response = t.type({
  sessionId: NonEmptyString,
  commonKeychain: NonEmptyString,
  bitgoMsg4: MPCv2BroadcastMessage,
});

export type MPCv2KeyGenRound3Response = t.TypeOf<
  typeof MPCv2KeyGenRound3Response
>;
