// @bitgo/sdk-coin-bera/dist/src/beraToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeraToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class BeraToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, BeraToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, BeraToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(BeraToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc **/
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /**
     * Make a query to Berachain explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response Berachain explorer
     */
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].beraExplorerApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].beraExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    getFullName() {
        return 'Bera Token';
    }
}
exports.BeraToken = BeraToken;
BeraToken.coinNames = {
    Mainnet: 'bera',
    Testnet: 'tbera',
};
