// @bitgo/public-types/src/schema/transactionRequest/message.ts
import * as t from "io-ts";
import { DateFromISOString } from "io-ts-types";
import { TransactionState } from "./transactionState";
import { SignatureShare } from "./signatureShare";
import { CommitmentShare } from "./commitmentShare";
import { EncryptedShare } from "./encryptedShare";

/**
 * Schema for the Message object used in TransactionRequests
 * Used for message signing
 */
export const Message = t.intersection([
  // Required fields
  t.type({
    /**
     * The current state of the message
     */
    state: TransactionState,

    /**
     * The raw message to be signed
     */
    messageRaw: t.string,
  }),
  // Optional fields
  t.partial({
    /**
     * Signature shares for the message
     */
    signatureShares: t.array(SignatureShare),

    /**
     * Private signature shares for the message
     */
    privateSignatureShares: t.array(SignatureShare),

    /**
     * Commitment shares for the message (used for EdDSA)
     */
    commitmentShares: t.array(CommitmentShare),

    /**
     * Encrypted shares for the message
     */
    encryptedShares: t.array(EncryptedShare),

    /**
     * Transaction hash
     */
    txHash: t.string,

    /**
     * Date when the message was last updated
     */
    updatedDate: DateFromISOString,

    /**
     * Date when the message was created
     */
    createdDate: DateFromISOString,

    /**
     * The encoded message
     */
    messageEncoded: t.string,

    /**
     * The broadcastable message
     */
    messageBroadcastable: t.string,

    /**
     * The encoded message
     */
    messageStandardType: t.string,

    /**
     * Derivation path for the message
     */
    derivationPath: t.string,

    /**
     * Combined signature share
     */
    combineSigShare: t.string,

    /**
     * Coin type for the message
     */
    coin: t.string,

    /**
     * BitGo's Paillier challenge for ECDSA
     */
    bitgoPaillierChallenge: t.unknown,

    /**
     * BitGo's encrypted challenge for ECDSA with Range Proofs
     */
    encryptedChallenge: t.string,
  }),
]);

export type Message = t.TypeOf<typeof Message>;
