// @bitgo/public-types/src/schema/transactionRequest/transactionRequestState.ts
import * as t from "io-ts";

/**
 * If supplied, only transaction request state changes from the provided list
 * will trigger notifications. If not provided, all transaction request state
 * changes will trigger notifications.
 */
export const TransactionRequestState = t.keyof(
  {
    pendingApproval: 1,
    canceled: 1,
    rejected: 1,
    /** @private */
    initialized: 1,
    pendingDelivery: 1,
    delivered: 1,
    pendingUserSignature: 1,
    pendingUserCommitment: 1,
    pendingUserRShare: 1,
    pendingUserGShare: 1,
    readyToSend: 1,
    signed: 1,
    failed: 1,
  },
  "TransactionRequestState",
);
export type TransactionRequestState = t.TypeOf<typeof TransactionRequestState>;

export const PendingTranscationRequestStates: readonly TransactionRequestState[] =
  [
    "pendingApproval",
    "pendingDelivery",
    "pendingUserSignature",
    "pendingUserCommitment",
    "pendingUserRShare",
    "pendingUserGShare",
    "readyToSend",
  ] as const;

export type PendingTranscationRequestState =
  (typeof PendingTranscationRequestStates)[number];
