// @bitgo/public-types/src/schema/transactionRequest/unsignedTransaction.ts
import * as t from "io-ts";

const Input = t.intersection([
  t.type({
    valueString: t.string,
  }),
  t.partial({
    _id: t.string,
    address: t.string,
    chain: t.number,
    derivationIndex: t.number,
    index: t.number,
    value: t.union([t.number, t.bigint]),
  }),
]);

const Output = t.intersection([
  t.type({
    valueString: t.string,
  }),
  t.partial({
    address: t.string,
    change: t.boolean,
    coinName: t.string,
    data: t.string,
    isPayGo: t.boolean,
    value: t.union([t.number, t.bigint]),
    wallet: t.string,
    walletV1: t.string,
    baseAddress: t.string,
    enterprise: t.string,
  }),
]);

const ParsedTx = t.intersection([
  t.type({
    inputs: t.array(Input),
    minerFee: t.union([t.string, t.number]),
    outputs: t.array(Output),
    spendAmount: t.union([t.string, t.number]),
  }),
  t.partial({
    gasPrice: t.number,
    hasUnvalidatedData: t.boolean,
    payGoFee: t.union([t.string, t.number]),
    spendAmounts: t.array(t.UnknownRecord),
    type: t.string,
  }),
]);

export const UnsignedTransaction = t.intersection([
  t.type({
    parsedTx: ParsedTx,
    /** The unsigned transaction as a hex string. */
    serializedTxHex: t.string,
    /** The portion of a transaction used to generate a signature (may or may not be the same as serializedTxHex) as a hex string. */
    signableHex: t.string,
  }),
  t.partial({
    /** A bip32 path. */
    derivationPath: t.string,
    /** Coin specific information. */
    coinSpecific: t.intersection([
      t.UnknownRecord,
      t.partial({
        nonceAddress: t.string,
        stakingAddress: t.string,
        signature: t.string,
      }),
    ]),
    feeInfo: t.type({
      fee: t.union([t.string, t.number]),
      feeString: t.string,
    }),
  }),
]);
export type UnsignedTransaction = t.TypeOf<typeof UnsignedTransaction>;
