// @bitgo/sdk-coin-sol/dist/cjs/test/unit/instructionParamsFactory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const testData = __importStar(require("../resources/sol"));
const instructionParamsFactory_1 = require("../../src/lib/instructionParamsFactory");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../../src/lib/constants");
const web3_js_1 = require("@solana/web3.js");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const spl_token_1 = require("@solana/spl-token");
describe('Instruction Parser Tests: ', function () {
    describe('Succeed ', function () {
        it('Wallet init tx instructions', () => {
            const fromAddress = testData.authAccount.pub;
            const nonceAddress = testData.nonceAccount.pub;
            const authAddress = testData.authAccount.pub;
            const amount = '100000';
            const instructions = web3_js_1.SystemProgram.createNonceAccount({
                fromPubkey: new web3_js_1.PublicKey(fromAddress),
                noncePubkey: new web3_js_1.PublicKey(nonceAddress),
                authorizedPubkey: new web3_js_1.PublicKey(authAddress),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const createNonceAccount = {
                type: constants_1.InstructionBuilderTypes.CreateNonceAccount,
                params: { fromAddress, nonceAddress, authAddress, amount },
            };
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.WalletInitialization, instructions);
            should_1.default.deepEqual(result, [createNonceAccount]);
        });
        it('Send tx instructions', () => {
            const authAccount = testData.authAccount.pub;
            const nonceAccount = testData.nonceAccount.pub;
            const amount = '100000';
            const memo = 'test memo';
            // nonce
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: authAccount },
            };
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: new web3_js_1.PublicKey(authAccount),
            });
            // transfer
            const transferParams = {
                type: constants_1.InstructionBuilderTypes.Transfer,
                params: { fromAddress: authAccount, toAddress: nonceAccount, amount },
            };
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount),
                toPubkey: new web3_js_1.PublicKey(nonceAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            });
            // memo
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            const instructions = [nonceAdvanceInstruction, transferInstruction, memoInstruction];
            const instructionsData = [nonceAdvanceParams, transferParams, memoParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Send token tx instructions', () => {
            const authAccount = testData.authAccount.pub;
            const nonceAccount = testData.nonceAccount.pub;
            const amount = testData.tokenTransfers.amount;
            const memo = testData.tokenTransfers.memo;
            const decimals = testData.tokenTransfers.decimals;
            const nameUSDC = testData.tokenTransfers.nameUSDC;
            const mintUSDC = testData.tokenTransfers.mintUSDC;
            const owner = testData.tokenTransfers.owner;
            const sourceUSDC = testData.tokenTransfers.sourceUSDC;
            // nonce
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: authAccount },
            };
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: new web3_js_1.PublicKey(authAccount),
            });
            // token transfer
            const transferParams = {
                type: constants_1.InstructionBuilderTypes.TokenTransfer,
                params: {
                    fromAddress: owner,
                    toAddress: nonceAccount,
                    amount: amount.toString(),
                    tokenName: nameUSDC,
                    sourceAddress: sourceUSDC,
                    programId: 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
                    tokenAddress: 'F4uLeXJoFz3hw13MposuwaQbMcZbCjqvEGPPeRRB1Byf',
                    decimalPlaces: 9,
                },
            };
            const transferInstruction = (0, spl_token_1.createTransferCheckedInstruction)(new web3_js_1.PublicKey(sourceUSDC), new web3_js_1.PublicKey(mintUSDC), new web3_js_1.PublicKey(nonceAccount), new web3_js_1.PublicKey(owner), amount, decimals);
            // memo
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            const instructions = [nonceAdvanceInstruction, transferInstruction, memoInstruction];
            const instructionsData = [nonceAdvanceParams, transferParams, memoParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Send sol 2022 token tx instructions', () => {
            const authAccount = testData.authAccount.pub;
            const nonceAccount = testData.nonceAccount.pub;
            const amount = testData.sol2022TokenTransfers.amount;
            const memo = testData.sol2022TokenTransfers.memo;
            const decimals = testData.sol2022TokenTransfers.decimals;
            const name = testData.sol2022TokenTransfers.name;
            const mint = testData.sol2022TokenTransfers.mint;
            const owner = testData.sol2022TokenTransfers.owner;
            const source = testData.sol2022TokenTransfers.source;
            // nonce
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: authAccount },
            };
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: new web3_js_1.PublicKey(authAccount),
            });
            // token transfer
            const transferParams = {
                type: constants_1.InstructionBuilderTypes.TokenTransfer,
                params: {
                    fromAddress: owner,
                    toAddress: nonceAccount,
                    amount: amount.toString(),
                    tokenName: name,
                    sourceAddress: source,
                    tokenAddress: '5NR1bQwLWqjbkhbQ1hx72HKJybbuvwkDnUZNoAZ2VhW6',
                    programId: 'TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb',
                    decimalPlaces: 9,
                },
            };
            const transferInstruction = (0, spl_token_1.createTransferCheckedInstruction)(new web3_js_1.PublicKey(source), new web3_js_1.PublicKey(mint), new web3_js_1.PublicKey(nonceAccount), new web3_js_1.PublicKey(owner), amount, decimals, [], spl_token_1.TOKEN_2022_PROGRAM_ID);
            // memo
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            const instructions = [nonceAdvanceInstruction, transferInstruction, memoInstruction];
            const instructionsData = [nonceAdvanceParams, transferParams, memoParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('multi ATA init tx instructions', () => {
            const ataParams = [
                {
                    mintAddress: testData.associatedTokenAccounts.mintId,
                    ownerAddress: testData.associatedTokenAccounts.accounts[0].pub,
                    payerAddress: testData.associatedTokenAccounts.accounts[0].pub,
                    ataAddress: testData.associatedTokenAccounts.accounts[0].ata,
                    programId: 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL',
                },
                {
                    mintAddress: testData.associatedTokenAccounts.mintId,
                    ownerAddress: testData.associatedTokenAccounts.accounts[1].pub,
                    payerAddress: testData.associatedTokenAccounts.accounts[0].pub,
                    ataAddress: testData.associatedTokenAccounts.accounts[1].ata,
                    programId: 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL',
                },
            ];
            const ataInstructions = [];
            const createATAParams = [];
            ataParams.forEach((param) => {
                ataInstructions.push((0, spl_token_1.createAssociatedTokenAccountInstruction)(new web3_js_1.PublicKey(param.payerAddress), new web3_js_1.PublicKey(param.ataAddress), new web3_js_1.PublicKey(param.ownerAddress), new web3_js_1.PublicKey(param.mintAddress)));
                createATAParams.push({
                    type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                    params: { ...param, tokenName: 'sol:usdc' },
                });
            });
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.AssociatedTokenAccountInitialization, ataInstructions);
            should_1.default.deepEqual(result, createATAParams);
        });
        it('sol 2022 ATA init tx instructions', () => {
            const ataParams = [
                {
                    mintAddress: testData.associatedTokenAccountsForSol2022.mintId,
                    ownerAddress: testData.associatedTokenAccountsForSol2022.accounts[0].pub,
                    payerAddress: testData.associatedTokenAccountsForSol2022.accounts[0].pub,
                    ataAddress: testData.associatedTokenAccountsForSol2022.accounts[0].ata,
                    programId: 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL',
                },
            ];
            const ataInstructions = [];
            const createATAParams = [];
            ataParams.forEach((param) => {
                ataInstructions.push((0, spl_token_1.createAssociatedTokenAccountInstruction)(new web3_js_1.PublicKey(param.payerAddress), new web3_js_1.PublicKey(param.ataAddress), new web3_js_1.PublicKey(param.ownerAddress), new web3_js_1.PublicKey(param.mintAddress), spl_token_1.TOKEN_2022_PROGRAM_ID));
                createATAParams.push({
                    type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                    params: { ...param, tokenName: 'tsol:t22mint' },
                });
            });
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.AssociatedTokenAccountInitialization, ataInstructions);
            should_1.default.deepEqual(result, createATAParams);
        });
        it('Mint To - Standard SPL Token instruction parsing', () => {
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const destinationAddress = testData.tokenTransfers.sourceUSDC;
            const authorityAddress = testData.authAccount.pub;
            const amount = '1000000';
            const tokenName = testData.tokenTransfers.nameUSDC;
            const mintInstruction = (0, spl_token_1.createMintToInstruction)(new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount));
            const expectedMintParams = {
                type: constants_1.InstructionBuilderTypes.MintTo,
                params: {
                    mintAddress,
                    destinationAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces: undefined,
                    programId: 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
                },
            };
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, [mintInstruction]);
            should_1.default.deepEqual(result, [expectedMintParams]);
        });
        it('Mint To - Token-2022 Program instruction parsing', () => {
            const mintAddress = testData.sol2022TokenTransfers.mint;
            const destinationAddress = testData.sol2022TokenTransfers.source;
            const authorityAddress = testData.authAccount.pub;
            const amount = '2000000';
            const tokenName = testData.sol2022TokenTransfers.name;
            const mintInstruction = (0, spl_token_1.createMintToInstruction)(new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount), undefined, spl_token_1.TOKEN_2022_PROGRAM_ID);
            const expectedMintParams = {
                type: constants_1.InstructionBuilderTypes.MintTo,
                params: {
                    mintAddress,
                    destinationAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces: undefined,
                    programId: spl_token_1.TOKEN_2022_PROGRAM_ID.toString(),
                },
            };
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, [mintInstruction]);
            should_1.default.deepEqual(result, [expectedMintParams]);
        });
        it('Burn - Standard SPL Token instruction parsing', () => {
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const accountAddress = testData.tokenTransfers.sourceUSDC;
            const authorityAddress = testData.authAccount.pub;
            const amount = '500000';
            const tokenName = testData.tokenTransfers.nameUSDC;
            const burnInstruction = (0, spl_token_1.createBurnInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount));
            const expectedBurnParams = {
                type: constants_1.InstructionBuilderTypes.Burn,
                params: {
                    mintAddress,
                    accountAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces: undefined,
                    programId: 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
                },
            };
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, [burnInstruction]);
            should_1.default.deepEqual(result, [expectedBurnParams]);
        });
        it('Burn - Token-2022 Program instruction parsing', () => {
            const mintAddress = testData.sol2022TokenTransfers.mint;
            const accountAddress = testData.sol2022TokenTransfers.source;
            const authorityAddress = testData.authAccount.pub;
            const amount = '750000';
            const tokenName = testData.sol2022TokenTransfers.name;
            const burnInstruction = (0, spl_token_1.createBurnInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount), undefined, spl_token_1.TOKEN_2022_PROGRAM_ID);
            const expectedBurnParams = {
                type: constants_1.InstructionBuilderTypes.Burn,
                params: {
                    mintAddress,
                    accountAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces: undefined,
                    programId: spl_token_1.TOKEN_2022_PROGRAM_ID.toString(),
                },
            };
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, [burnInstruction]);
            should_1.default.deepEqual(result, [expectedBurnParams]);
        });
        it('Mixed instructions - Mint, Burn, and Transfer', () => {
            const authAccount = testData.authAccount.pub;
            const nonceAccount = testData.nonceAccount.pub;
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const destinationAddress = testData.tokenTransfers.sourceUSDC;
            const accountAddress = testData.tokenTransfers.sourceUSDC;
            const tokenName = testData.tokenTransfers.nameUSDC;
            // Nonce advance
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: new web3_js_1.PublicKey(authAccount),
            });
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: authAccount },
            };
            // Mint instruction
            const mintInstruction = (0, spl_token_1.createMintToInstruction)(new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authAccount), BigInt('1000000'));
            const expectedMintParams = {
                type: constants_1.InstructionBuilderTypes.MintTo,
                params: {
                    mintAddress,
                    destinationAddress,
                    authorityAddress: authAccount,
                    amount: '1000000',
                    tokenName,
                    decimalPlaces: undefined,
                    programId: 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
                },
            };
            // Burn instruction
            const burnInstruction = (0, spl_token_1.createBurnInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(authAccount), BigInt('500000'));
            const expectedBurnParams = {
                type: constants_1.InstructionBuilderTypes.Burn,
                params: {
                    mintAddress,
                    accountAddress,
                    authorityAddress: authAccount,
                    amount: '500000',
                    tokenName,
                    decimalPlaces: undefined,
                    programId: 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
                },
            };
            const instructions = [nonceAdvanceInstruction, mintInstruction, burnInstruction];
            const expectedParams = [nonceAdvanceParams, expectedMintParams, expectedBurnParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, instructions);
            should_1.default.deepEqual(result, expectedParams);
        });
    });
    describe('Fail ', function () {
        it('Invalid type', () => {
            (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.ContractCall, [])).throwError('Invalid transaction, transaction type not supported: ' + sdk_core_1.TransactionType.ContractCall);
        });
        it('Invalid Instruction for Send Type', () => {
            const fromAddress = testData.authAccount.pub;
            const nonceAddress = testData.nonceAccount.pub;
            const authAddress = testData.authAccount.pub;
            const amount = '100000';
            const instructions = web3_js_1.SystemProgram.createNonceAccount({
                fromPubkey: new web3_js_1.PublicKey(fromAddress),
                noncePubkey: new web3_js_1.PublicKey(nonceAddress),
                authorizedPubkey: new web3_js_1.PublicKey(authAddress),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.Send, instructions)).throwError('Invalid transaction, instruction type not supported: Create');
        });
    });
});
