// @bitgo/sdk-coin-sol/dist/cjs/test/unit/instructionParamsFactory.staking.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const testData = __importStar(require("../resources/sol"));
const instructionParamsFactory_1 = require("../../src/lib/instructionParamsFactory");
const public_types_1 = require("@bitgo/public-types");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../../src/lib/constants");
const web3_js_1 = require("@solana/web3.js");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
describe('Instruction Parser Staking Tests: ', function () {
    describe('Activate staking instructions ', function () {
        it('Should parse activate stake tx instructions with memo and durable nonce', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const nonceAccount = testData.nonceAccount.pub;
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            const memo = 'test memo';
            // Instructions
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: fromAccount,
            });
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            // Params
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
            };
            const stakingActivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    validator: validator.toString(),
                    amount,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const instructions = [
                nonceAdvanceInstruction,
                ...stakingActivateInstructions,
                ...stakingDelegateInstructions,
                memoInstruction,
            ];
            const instructionsData = [nonceAdvanceParams, memoParams, stakingActivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingActivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse activate stake tx instructions with memo and durable nonce with instructions in any order', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const nonceAccount = testData.nonceAccount.pub;
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            const memo = 'test memo';
            // Instructions
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: fromAccount,
            });
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            // Params
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
            };
            const stakingActivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    validator: validator.toString(),
                    amount,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const instructions = [
                memoInstruction,
                ...stakingActivateInstructions,
                ...stakingDelegateInstructions,
                nonceAdvanceInstruction,
            ];
            const instructionsData = [memoParams, nonceAdvanceParams, stakingActivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingActivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse activate stake tx instructions without memo or durable nonce', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            // Instructions
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            // Params
            const stakingActivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    validator: validator.toString(),
                    amount,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const instructions = [...stakingActivateInstructions, ...stakingDelegateInstructions];
            const instructionsData = [stakingActivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingActivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse activate stake tx instructions if there are unexpected instructions', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            // Instructions
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
            }).instructions;
            // Params
            const stakingActivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    validator: validator.toString(),
                    amount,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const instructions = [
                ...stakingActivateInstructions,
                ...stakingDelegateInstructions,
                ...stakingDeactivateInstructions,
            ];
            const instructionsData = [stakingActivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingActivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should fail to parse activate stake tx instructions if there are missing instructions', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            // Instructions
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            should_1.default.throws(() => {
                const instructions = [stakingActivateInstructions[1], ...stakingDelegateInstructions];
                (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingActivate, instructions);
            }, 'Invalid staking activate transaction, missing create stake account instruction');
            should_1.default.throws(() => {
                const instructions = [stakingActivateInstructions[0]];
                (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingActivate, instructions);
            }, 'Invalid staking activate transaction, missing initialize stake account/delegate instruction');
        });
    });
    describe('Deactivate staking instructions ', function () {
        it('Should parse deactivate stake tx instructions with memo and durable nonce', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const nonceAccount = testData.nonceAccount.pub;
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const memo = 'test memo';
            // Instructions
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: fromAccount,
            });
            const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
            }).instructions;
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            // Params
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
            };
            const stakingDeactivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount: undefined,
                    unstakingAddress: undefined,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const instructions = [nonceAdvanceInstruction, ...stakingDeactivateInstructions, memoInstruction];
            const instructionsData = [nonceAdvanceParams, memoParams, stakingDeactivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse deactivate stake tx instructions with memo and durable nonce with instructions in any order', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const nonceAccount = testData.nonceAccount.pub;
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const memo = 'test memo';
            // Instructions
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: fromAccount,
            });
            const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
            }).instructions;
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            // Params
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
            };
            const stakingDeactivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount: undefined,
                    unstakingAddress: undefined,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const instructions = [memoInstruction, ...stakingDeactivateInstructions, nonceAdvanceInstruction];
            const instructionsData = [memoParams, nonceAdvanceParams, stakingDeactivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse deactivate stake tx instructions without memo or durable nonce', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            // Instructions
            const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
            }).instructions;
            // Params
            const stakingDeactivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount: undefined,
                    unstakingAddress: undefined,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const instructions = [...stakingDeactivateInstructions];
            const instructionsData = [stakingDeactivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse deactivate stake tx instructions if there are unexpected instructions', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            // Instructions
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
            }).instructions;
            // Params
            const stakingActivateParams = {
                type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount: undefined,
                    unstakingAddress: undefined,
                    stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                },
            };
            const instructions = [
                ...stakingActivateInstructions,
                ...stakingDelegateInstructions,
                ...stakingDeactivateInstructions,
            ];
            const instructionsData = [stakingActivateParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        describe('Partially deactivate stake instructions', function () {
            describe('Input validation', function () {
                it('Should throw an error if the Allocate instruction is missing', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: splitStakeAccount,
                        programId: web3_js_1.StakeProgram.programId,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: splitStakeAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        assignInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, missing allocate unstake account instruction');
                });
                it('Should throw an error if the Assign instruction is missing', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: splitStakeAccount,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: splitStakeAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, missing assign unstake account instruction');
                });
                it('Should throw an error if the Split instruction is missing', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: splitStakeAccount,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: splitStakeAccount,
                        programId: web3_js_1.StakeProgram.programId,
                    });
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        assignInstruction,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, missing split stake account instruction');
                });
                it('Should throw an error if the transfer instruction is missing for partial', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: splitStakeAccount,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: splitStakeAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: splitStakeAccount,
                        programId: web3_js_1.StakeProgram.programId,
                    });
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        assignInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, missing funding of unstake address instruction');
                });
                it('Should throw an error if the allocated account does not match the assigned account', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: new web3_js_1.Keypair().publicKey,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: splitStakeAccount,
                        programId: web3_js_1.StakeProgram.programId,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: splitStakeAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        assignInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, must allocate and assign the same public key');
                });
                [199, 201].forEach((space) => {
                    it(`Should throw an error if the correct amount of space is not allocated for the split account - ${space}`, () => {
                        const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                        const nonceAccount = testData.nonceAccount.pub;
                        const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                        const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                        const memo = 'test memo';
                        // Instructions
                        const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                            noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                            authorizedPubkey: fromAccount,
                        });
                        const allocateInstruction = web3_js_1.SystemProgram.allocate({
                            accountPubkey: splitStakeAccount,
                            space,
                        });
                        const assignInstruction = web3_js_1.SystemProgram.assign({
                            accountPubkey: splitStakeAccount,
                            programId: web3_js_1.StakeProgram.programId,
                        });
                        const splitInstructions = web3_js_1.StakeProgram.split({
                            stakePubkey: stakingAccount,
                            authorizedPubkey: fromAccount,
                            splitStakePubkey: splitStakeAccount,
                            lamports: 100000,
                        }, 0).instructions;
                        const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                            authorizedPubkey: fromAccount,
                            stakePubkey: splitStakeAccount,
                        }).instructions;
                        const memoInstruction = new web3_js_1.TransactionInstruction({
                            keys: [],
                            programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                            data: Buffer.from(memo),
                        });
                        const instructions = [
                            nonceAdvanceInstruction,
                            allocateInstruction,
                            assignInstruction,
                            ...splitInstructions,
                            ...stakingDeactivateInstructions,
                            memoInstruction,
                        ];
                        (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw(`Invalid partial deactivate stake transaction, unstaking account must allocate ${web3_js_1.StakeProgram.space} bytes`);
                    });
                });
                it('Should throw an error if the allocated account is not assigned to the StakeProgram', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: splitStakeAccount,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: splitStakeAccount,
                        programId: web3_js_1.SystemProgram.programId,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: splitStakeAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        assignInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, the unstake account must be assigned to the Stake Program');
                });
                it('Should throw an error if the split account is not allocated', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const key = new web3_js_1.Keypair().publicKey;
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: key,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: key,
                        programId: web3_js_1.StakeProgram.programId,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: stakingAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        assignInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, must allocate the unstaking account');
                });
                it('Should throw an error if the stake account and the split account are the same account', () => {
                    const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                    const nonceAccount = testData.nonceAccount.pub;
                    const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                    const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                    const memo = 'test memo';
                    // Instructions
                    const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                        noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                        authorizedPubkey: fromAccount,
                    });
                    const allocateInstruction = web3_js_1.SystemProgram.allocate({
                        accountPubkey: stakingAccount,
                        space: web3_js_1.StakeProgram.space,
                    });
                    const assignInstruction = web3_js_1.SystemProgram.assign({
                        accountPubkey: stakingAccount,
                        programId: web3_js_1.StakeProgram.programId,
                    });
                    const splitInstructions = web3_js_1.StakeProgram.split({
                        stakePubkey: stakingAccount,
                        authorizedPubkey: fromAccount,
                        splitStakePubkey: stakingAccount,
                        lamports: 100000,
                    }, 0).instructions;
                    const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                        authorizedPubkey: fromAccount,
                        stakePubkey: splitStakeAccount,
                    }).instructions;
                    const memoInstruction = new web3_js_1.TransactionInstruction({
                        keys: [],
                        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                        data: Buffer.from(memo),
                    });
                    const instructions = [
                        nonceAdvanceInstruction,
                        allocateInstruction,
                        assignInstruction,
                        ...splitInstructions,
                        ...stakingDeactivateInstructions,
                        memoInstruction,
                    ];
                    (0, should_1.default)(() => (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions)).throw('Invalid partial deactivate stake transaction, the unstaking account must be different from the Stake Account');
                });
            });
            it('Should parse partial deactivate stake tx instructions with memo and durable nonce', () => {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const nonceAccount = testData.nonceAccount.pub;
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                const splitStakeAccount = new web3_js_1.PublicKey(testData.splitStakeAccount.pub);
                const memo = 'test memo';
                // Instructions
                const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                    noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                    authorizedPubkey: fromAccount,
                });
                // transfer
                const transferInstruction = web3_js_1.SystemProgram.transfer({
                    fromPubkey: new web3_js_1.PublicKey(fromAccount),
                    toPubkey: new web3_js_1.PublicKey(splitStakeAccount),
                    lamports: parseInt(constants_1.STAKE_ACCOUNT_RENT_EXEMPT_AMOUNT.toString(), 10),
                });
                const allocateInstruction = web3_js_1.SystemProgram.allocate({
                    accountPubkey: splitStakeAccount,
                    space: web3_js_1.StakeProgram.space,
                });
                const assignInstruction = web3_js_1.SystemProgram.assign({
                    accountPubkey: splitStakeAccount,
                    programId: web3_js_1.StakeProgram.programId,
                });
                const splitInstructions = web3_js_1.StakeProgram.split({
                    stakePubkey: stakingAccount,
                    authorizedPubkey: fromAccount,
                    splitStakePubkey: splitStakeAccount,
                    lamports: 100000,
                }, 0).instructions;
                const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                    authorizedPubkey: fromAccount,
                    stakePubkey: splitStakeAccount,
                }).instructions;
                const memoInstruction = new web3_js_1.TransactionInstruction({
                    keys: [],
                    programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                    data: Buffer.from(memo),
                });
                // Params
                const nonceAdvanceParams = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
                };
                const stakingDeactivateParams = {
                    type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                    params: {
                        fromAddress: fromAccount.toString(),
                        stakingAddress: stakingAccount.toString(),
                        amount: '100000',
                        unstakingAddress: splitStakeAccount.toString(),
                        stakingType: public_types_1.SolStakingTypeEnum.NATIVE,
                    },
                };
                const memoParams = {
                    type: constants_1.InstructionBuilderTypes.Memo,
                    params: { memo },
                };
                const instructions = [
                    nonceAdvanceInstruction,
                    transferInstruction,
                    allocateInstruction,
                    assignInstruction,
                    ...splitInstructions,
                    ...stakingDeactivateInstructions,
                    memoInstruction,
                ];
                const instructionsData = [nonceAdvanceParams, memoParams, stakingDeactivateParams];
                const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingDeactivate, instructions);
                should_1.default.deepEqual(result, instructionsData);
            });
        });
    });
    describe('Withdraw stake instructions ', function () {
        it('Should parse withdraw stake tx instructions with memo and durable nonce', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const nonceAccount = testData.nonceAccount.pub;
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const memo = 'test memo';
            const amount = '100000';
            // Instructions
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: fromAccount,
            });
            const withdrawStakeInstructions = web3_js_1.StakeProgram.withdraw({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                toPubkey: fromAccount,
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            // Params
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
            };
            const withdrawStakeParams = {
                type: constants_1.InstructionBuilderTypes.StakingWithdraw,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount,
                },
            };
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const instructions = [nonceAdvanceInstruction, ...withdrawStakeInstructions, memoInstruction];
            const instructionsData = [nonceAdvanceParams, withdrawStakeParams, memoParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingWithdraw, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse withdraw stake tx instructions with memo and durable nonce with instructions in any order', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const nonceAccount = testData.nonceAccount.pub;
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const memo = 'test memo';
            const amount = '100000';
            // Instructions
            const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                authorizedPubkey: fromAccount,
            });
            const withdrawStakeInstructions = web3_js_1.StakeProgram.withdraw({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                toPubkey: fromAccount,
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from(memo),
            });
            // Params
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress: nonceAccount, authWalletAddress: fromAccount.toString() },
            };
            const withdrawStakeParams = {
                type: constants_1.InstructionBuilderTypes.StakingWithdraw,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount,
                },
            };
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const instructions = [memoInstruction, ...withdrawStakeInstructions, nonceAdvanceInstruction];
            const instructionsData = [memoParams, withdrawStakeParams, nonceAdvanceParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingWithdraw, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse withdraw stake tx instructions without memo or durable nonce', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const amount = '100000';
            // Instructions
            const withdrawStakeInstructions = web3_js_1.StakeProgram.withdraw({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                toPubkey: fromAccount,
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            // Params
            const withdrawStakeParams = {
                type: constants_1.InstructionBuilderTypes.StakingWithdraw,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount,
                },
            };
            const instructions = [...withdrawStakeInstructions];
            const instructionsData = [withdrawStakeParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingWithdraw, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
        it('Should parse withdraw stake tx instructions if there are unexpected instructions', () => {
            const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
            const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
            const validator = new web3_js_1.PublicKey(testData.validator.pub);
            const amount = '100000';
            // Instructions
            const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                fromPubkey: fromAccount,
                stakePubkey: stakingAccount,
                authorized: {
                    staker: fromAccount,
                    withdrawer: fromAccount,
                },
                lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                votePubkey: validator,
            }).instructions;
            const withdrawStakeInstructions = web3_js_1.StakeProgram.withdraw({
                authorizedPubkey: fromAccount,
                stakePubkey: stakingAccount,
                toPubkey: fromAccount,
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions;
            // Params
            const withdrawStakeParams = {
                type: constants_1.InstructionBuilderTypes.StakingWithdraw,
                params: {
                    fromAddress: fromAccount.toString(),
                    stakingAddress: stakingAccount.toString(),
                    amount,
                },
            };
            const instructions = [
                ...stakingActivateInstructions,
                ...stakingDelegateInstructions,
                ...withdrawStakeInstructions,
            ];
            const instructionsData = [withdrawStakeParams];
            const result = (0, instructionParamsFactory_1.instructionParamsFactory)(sdk_core_1.TransactionType.StakingWithdraw, instructions);
            should_1.default.deepEqual(result, instructionsData);
        });
    });
});
