// @bitgo/sdk-coin-sol/dist/cjs/test/unit/solInstructionFactory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const testData = __importStar(require("../resources/sol"));
const solInstructionFactory_1 = require("../../src/lib/solInstructionFactory");
const token2022Config_1 = require("../../src/lib/token2022Config");
const constants_1 = require("../../src/lib/constants");
const web3_js_1 = require("@solana/web3.js");
const spl_token_1 = require("@solana/spl-token");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
describe('Instruction Builder Tests: ', function () {
    describe('Succeed ', function () {
        it('Memo', () => {
            const memo = 'test memo';
            const memoParams = {
                type: constants_1.InstructionBuilderTypes.Memo,
                params: { memo },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(memoParams);
            should_1.default.deepEqual(result, [
                new web3_js_1.TransactionInstruction({
                    keys: [],
                    programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                    data: Buffer.from(memo),
                }),
            ]);
        });
        it('Transfer', () => {
            const fromAddress = testData.authAccount.pub;
            const toAddress = testData.nonceAccount.pub;
            const amount = '100000';
            const transferParams = {
                type: constants_1.InstructionBuilderTypes.Transfer,
                params: { fromAddress, toAddress, amount },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(transferParams);
            should_1.default.deepEqual(result, [
                web3_js_1.SystemProgram.transfer({
                    fromPubkey: new web3_js_1.PublicKey(fromAddress),
                    toPubkey: new web3_js_1.PublicKey(toAddress),
                    lamports: new bignumber_js_1.default(amount).toNumber(),
                }),
            ]);
        });
        it('Advance nonce', () => {
            const authWalletAddress = testData.authAccount.pub;
            const walletNonceAddress = testData.nonceAccount.pub;
            const nonceAdvanceParams = {
                type: constants_1.InstructionBuilderTypes.NonceAdvance,
                params: { walletNonceAddress, authWalletAddress },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(nonceAdvanceParams);
            should_1.default.deepEqual(result, [
                web3_js_1.SystemProgram.nonceAdvance({
                    noncePubkey: new web3_js_1.PublicKey(walletNonceAddress),
                    authorizedPubkey: new web3_js_1.PublicKey(authWalletAddress),
                }),
            ]);
        });
        it('Create and Nonce initialize', () => {
            const fromAddress = testData.authAccount.pub;
            const nonceAddress = testData.nonceAccount.pub;
            const authAddress = testData.authAccount.pub;
            const amount = '100000';
            const createNonceAccountParams = {
                type: constants_1.InstructionBuilderTypes.CreateNonceAccount,
                params: { fromAddress, nonceAddress, authAddress, amount },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(createNonceAccountParams);
            should_1.default.deepEqual(result, web3_js_1.SystemProgram.createNonceAccount({
                fromPubkey: new web3_js_1.PublicKey(fromAddress),
                noncePubkey: new web3_js_1.PublicKey(nonceAddress),
                authorizedPubkey: new web3_js_1.PublicKey(authAddress),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            }).instructions);
        });
        it('Create associated token account', () => {
            const mintAddress = testData.associatedTokenAccounts.mintId;
            const ataAddress = testData.associatedTokenAccounts.accounts[0].ata;
            const ownerAddress = testData.associatedTokenAccounts.accounts[0].pub;
            const payerAddress = testData.associatedTokenAccounts.accounts[0].pub;
            const createATAParams = {
                type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                params: {
                    mintAddress,
                    ataAddress,
                    ownerAddress,
                    payerAddress,
                    tokenName: testData.associatedTokenAccounts.mint,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(createATAParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createAssociatedTokenAccountInstruction)(new web3_js_1.PublicKey(payerAddress), new web3_js_1.PublicKey(ataAddress), new web3_js_1.PublicKey(ownerAddress), new web3_js_1.PublicKey(mintAddress)),
            ]);
        });
        it('Token Transfer', () => {
            const fromAddress = testData.authAccount.pub;
            const toAddress = testData.nonceAccount.pub;
            const amount = '100000';
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const sourceAddress = testData.associatedTokenAccounts.accounts[0].pub;
            const coinName = testData.tokenTransfers.nameUSDC;
            const transferParams = {
                type: constants_1.InstructionBuilderTypes.TokenTransfer,
                params: {
                    fromAddress: fromAddress,
                    toAddress: toAddress,
                    amount: amount,
                    tokenName: coinName,
                    sourceAddress: sourceAddress,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(transferParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createTransferCheckedInstruction)(new web3_js_1.PublicKey(sourceAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(toAddress), new web3_js_1.PublicKey(fromAddress), BigInt(amount), 9),
            ]);
        });
        it('Token Transfer - Token-2022 with transfer hook config', () => {
            const tokenConfig = (0, token2022Config_1.getToken2022Config)('4MmJVdwYN8LwvbGeCowYjSx7KoEi6BJWg8XXnW4fDDp6');
            should_1.default.exist(tokenConfig);
            should_1.default.exist(tokenConfig?.transferHook);
            const transferHook = tokenConfig.transferHook;
            const fromAddress = testData.authAccount.pub;
            const toAddress = testData.nonceAccount.pub;
            const sourceAddress = testData.associatedTokenAccounts.accounts[0].ata;
            const amount = '500000';
            const transferParams = {
                type: constants_1.InstructionBuilderTypes.TokenTransfer,
                params: {
                    fromAddress,
                    toAddress,
                    amount,
                    tokenName: 'tbill',
                    sourceAddress,
                    tokenAddress: tokenConfig.mintAddress,
                    decimalPlaces: 6,
                    programId: spl_token_1.TOKEN_2022_PROGRAM_ID.toString(),
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(transferParams);
            result.should.have.length(1);
            const builtInstruction = result[0];
            builtInstruction.programId.equals(spl_token_1.TOKEN_2022_PROGRAM_ID).should.be.true();
            const baseInstruction = (0, spl_token_1.createTransferCheckedInstruction)(new web3_js_1.PublicKey(sourceAddress), new web3_js_1.PublicKey(tokenConfig.mintAddress), new web3_js_1.PublicKey(toAddress), new web3_js_1.PublicKey(fromAddress), BigInt(amount), 6, [], spl_token_1.TOKEN_2022_PROGRAM_ID);
            const baseKeyCount = baseInstruction.keys.length;
            builtInstruction.keys.slice(0, baseKeyCount).should.deepEqual(baseInstruction.keys);
            const extraKeys = builtInstruction.keys.slice(baseKeyCount);
            const expectedExtraKeys = [
                ...transferHook.extraAccountMetas.map((meta) => ({
                    pubkey: new web3_js_1.PublicKey(meta.pubkey),
                    isSigner: meta.isSigner,
                    isWritable: meta.isWritable,
                })),
            ];
            extraKeys.should.deepEqual(expectedExtraKeys);
            for (const expectedMeta of expectedExtraKeys) {
                builtInstruction.keys.filter((meta) => meta.pubkey.equals(expectedMeta.pubkey)).should.have.length(1);
            }
        });
        it('Mint To - Standard SPL Token', () => {
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const destinationAddress = testData.tokenTransfers.sourceUSDC;
            const authorityAddress = testData.authAccount.pub;
            const amount = '1000000';
            const tokenName = testData.tokenTransfers.nameUSDC;
            const decimalPlaces = testData.tokenTransfers.decimals;
            const mintParams = {
                type: constants_1.InstructionBuilderTypes.MintTo,
                params: {
                    mintAddress,
                    destinationAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(mintParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createMintToInstruction)(new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount)),
            ]);
        });
        it('Mint To - Token-2022 Program', () => {
            const mintAddress = testData.sol2022TokenTransfers.mint;
            const destinationAddress = testData.sol2022TokenTransfers.source;
            const authorityAddress = testData.authAccount.pub;
            const amount = '2000000';
            const tokenName = testData.sol2022TokenTransfers.name;
            const decimalPlaces = testData.sol2022TokenTransfers.decimals;
            const mintParams = {
                type: constants_1.InstructionBuilderTypes.MintTo,
                params: {
                    mintAddress,
                    destinationAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces,
                    programId: spl_token_1.TOKEN_2022_PROGRAM_ID.toString(),
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(mintParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createMintToInstruction)(new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount), undefined, spl_token_1.TOKEN_2022_PROGRAM_ID),
            ]);
        });
        it('Burn - Standard SPL Token', () => {
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const accountAddress = testData.tokenTransfers.sourceUSDC;
            const authorityAddress = testData.authAccount.pub;
            const amount = '500000';
            const tokenName = testData.tokenTransfers.nameUSDC;
            const decimalPlaces = testData.tokenTransfers.decimals;
            const burnParams = {
                type: constants_1.InstructionBuilderTypes.Burn,
                params: {
                    mintAddress,
                    accountAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(burnParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createBurnInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount)),
            ]);
        });
        it('Burn - Token-2022 Program', () => {
            const mintAddress = testData.sol2022TokenTransfers.mint;
            const accountAddress = testData.sol2022TokenTransfers.source;
            const authorityAddress = testData.authAccount.pub;
            const amount = '750000';
            const tokenName = testData.sol2022TokenTransfers.name;
            const decimalPlaces = testData.sol2022TokenTransfers.decimals;
            const burnParams = {
                type: constants_1.InstructionBuilderTypes.Burn,
                params: {
                    mintAddress,
                    accountAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                    decimalPlaces,
                    programId: spl_token_1.TOKEN_2022_PROGRAM_ID.toString(),
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(burnParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createBurnInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount), undefined, spl_token_1.TOKEN_2022_PROGRAM_ID),
            ]);
        });
        it('Mint To - Without decimal places', () => {
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const destinationAddress = testData.tokenTransfers.sourceUSDC;
            const authorityAddress = testData.authAccount.pub;
            const amount = '1000000';
            const tokenName = testData.tokenTransfers.nameUSDC;
            const mintParams = {
                type: constants_1.InstructionBuilderTypes.MintTo,
                params: {
                    mintAddress,
                    destinationAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(mintParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createMintToInstruction)(new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount)),
            ]);
        });
        it('Burn - Without decimal places', () => {
            const mintAddress = testData.tokenTransfers.mintUSDC;
            const accountAddress = testData.tokenTransfers.sourceUSDC;
            const authorityAddress = testData.authAccount.pub;
            const amount = '500000';
            const tokenName = testData.tokenTransfers.nameUSDC;
            const burnParams = {
                type: constants_1.InstructionBuilderTypes.Burn,
                params: {
                    mintAddress,
                    accountAddress,
                    authorityAddress,
                    amount,
                    tokenName,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(burnParams);
            should_1.default.deepEqual(result, [
                (0, spl_token_1.createBurnInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(authorityAddress), BigInt(amount)),
            ]);
        });
        it('Custom instruction with TSS format', () => {
            const tssInstruction = {
                programId: '11111111111111111111111111111112',
                keys: [
                    {
                        pubkey: 'CyjoLt3kjqB57K7ewCBHmnHq3UgEj3ak6A7m6EsBsuhA',
                        isSigner: true,
                        isWritable: true,
                    },
                ],
                data: 'dGVzdCBpbnN0cnVjdGlvbiBkYXRh', // "test instruction data" in base64
            };
            const customInstructionParams = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    ...tssInstruction,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(customInstructionParams);
            result.should.have.length(1);
            const resultInstruction = result[0];
            resultInstruction.programId.toString().should.equal('11111111111111111111111111111112');
            resultInstruction.keys.should.have.length(1);
            resultInstruction.keys[0].pubkey.toString().should.equal('CyjoLt3kjqB57K7ewCBHmnHq3UgEj3ak6A7m6EsBsuhA');
            resultInstruction.keys[0].isSigner.should.equal(true);
            resultInstruction.keys[0].isWritable.should.equal(true);
            resultInstruction.data.toString().should.equal('test instruction data');
        });
        it('Custom instruction with TSS format - hex data', () => {
            const tssInstruction = {
                programId: '11111111111111111111111111111112',
                keys: [
                    {
                        pubkey: 'CyjoLt3kjqB57K7ewCBHmnHq3UgEj3ak6A7m6EsBsuhA',
                        isSigner: false,
                        isWritable: false,
                    },
                ],
                data: '74657374', // "test" in hex
            };
            const customInstructionParams = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    ...tssInstruction,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(customInstructionParams);
            result.should.have.length(1);
            const resultInstruction = result[0];
            resultInstruction.programId.toString().should.equal('11111111111111111111111111111112');
            resultInstruction.keys.should.have.length(1);
            resultInstruction.keys[0].pubkey.toString().should.equal('CyjoLt3kjqB57K7ewCBHmnHq3UgEj3ak6A7m6EsBsuhA');
            resultInstruction.keys[0].isSigner.should.equal(false);
            resultInstruction.keys[0].isWritable.should.equal(false);
            // Note: hex data will fall back to base64 then UTF-8, so it should contain the hex string
            resultInstruction.data.should.be.instanceOf(Buffer);
        });
        it('Custom instruction with TSS format - UTF-8 fallback', () => {
            const tssInstruction = {
                programId: '11111111111111111111111111111112',
                keys: [],
                data: 'hello world', // plain text, should use UTF-8 encoding
            };
            const customInstructionParams = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    ...tssInstruction,
                },
            };
            const result = (0, solInstructionFactory_1.solInstructionFactory)(customInstructionParams);
            result.should.have.length(1);
            const resultInstruction = result[0];
            resultInstruction.programId.toString().should.equal('11111111111111111111111111111112');
            resultInstruction.keys.should.have.length(0);
            // Since "hello world" is not valid base64, it should fall back to UTF-8
            resultInstruction.data.toString('utf8').should.equal('hello world');
        });
    });
    describe('Fail ', function () {
        it('Invalid type', () => {
            // @ts-expect-error Testing for an invalid type, should throw error
            (0, should_1.default)(() => (0, solInstructionFactory_1.solInstructionFactory)({ type: 'random', params: {} })).throwError('Invalid instruction type or not supported');
        });
        it('Custom instruction - missing programId', () => {
            const customInstructionParams = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    keys: [],
                    data: 'test',
                },
            };
            (0, should_1.default)(() => (0, solInstructionFactory_1.solInstructionFactory)(customInstructionParams)).throwError('Missing programId in custom instruction');
        });
        it('Custom instruction - missing keys', () => {
            const customInstructionParams = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    programId: '11111111111111111111111111111112',
                    data: 'test',
                },
            };
            (0, should_1.default)(() => (0, solInstructionFactory_1.solInstructionFactory)(customInstructionParams)).throwError('Missing or invalid keys in custom instruction');
        });
        it('Custom instruction - missing data', () => {
            const customInstructionParams = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    programId: '11111111111111111111111111111112',
                    keys: [
                        {
                            pubkey: 'CyjoLt3kjqB57K7ewCBHmnHq3UgEj3ak6A7m6EsBsuhA',
                            isSigner: true,
                            isWritable: true,
                        },
                    ],
                },
            };
            (0, should_1.default)(() => (0, solInstructionFactory_1.solInstructionFactory)(customInstructionParams)).throwError('Missing data in custom instruction');
        });
    });
});
