// @bitgo/sdk-coin-sol/dist/cjs/test/unit/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const statics_1 = require("@bitgo/statics");
const lib_1 = require("../../src/lib");
const testData = __importStar(require("../resources/sol"));
const web3_js_1 = require("@solana/web3.js");
const getBuilderFactory_1 = require("./getBuilderFactory");
describe('Sol Transaction', () => {
    const coin = statics_1.coins.get('tsol');
    describe('toJson should', () => {
        it('throw empty transaction', () => {
            const tx = new lib_1.Transaction(coin);
            assert_1.default.throws(() => tx.toJson(), /Empty transaction/);
            assert_1.default.throws(() => tx.toBroadcastFormat(), /Empty transaction/);
        });
        it('throw for toJson of empty tx', () => {
            const tx = new lib_1.Transaction(coin);
            (0, should_1.default)(() => tx.toJson()).throwError('Empty transaction');
            tx.solTransaction = new web3_js_1.Transaction();
            tx.solTransaction.recentBlockhash = testData.blockHashes.validBlockHashes[0];
            (0, should_1.default)(() => tx.toJson()).throwError('Invalid transaction, transaction type not supported: undefined');
        });
        it('succeed for a unsigned transfer tx', () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.TRANSFER_UNSIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE);
            tx.signature.should.be.empty();
            const txJson = tx.toJson();
            txJson.should.have.properties(['id', 'feePayer', 'nonce', 'numSignatures', 'instructionsData']);
            txJson.id?.should.equal(undefined);
            txJson.feePayer?.should.equal('5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe');
            txJson.nonce.should.equal('GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi');
            txJson.numSignatures.should.equal(0);
            txJson.instructionsData.length.should.equal(3);
            txJson.lamportsPerSignature?.should.be.undefined();
            txJson.instructionsData.should.deepEqual([
                {
                    type: 'NonceAdvance',
                    params: {
                        walletNonceAddress: '8Y7RM6JfcX4ASSNBkrkrmSbRu431YVi9Y3oLFnzC2dCh',
                        authWalletAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: 'CP5Dpaa42RtJmMuKqCQsLwma5Yh3knuvKsYDFX85F41S',
                        amount: '300000',
                    },
                },
                {
                    type: 'Memo',
                    params: {
                        memo: 'test memo',
                    },
                },
            ]);
        });
        it('succeed for a multi transfer tx', () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.MULTI_TRANSFER_SIGNED);
            const txJson = tx.toJson();
            txJson.should.have.properties(['id', 'feePayer', 'nonce', 'numSignatures', 'instructionsData']);
            txJson.id?.should.equal('TPVcc18CYxPnM3eRgQhdb6V6ZLa34Dv3dU7MtvKPuy5ZPKLM1uZPFFEmF2m184PTWKRZ1Uq6NKFZWwr2krKk63f');
            tx.signature.should.deepEqual([
                'TPVcc18CYxPnM3eRgQhdb6V6ZLa34Dv3dU7MtvKPuy5ZPKLM1uZPFFEmF2m184PTWKRZ1Uq6NKFZWwr2krKk63f',
            ]);
            txJson.feePayer?.should.equal('5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe');
            txJson.nonce.should.equal('GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi');
            txJson.numSignatures.should.equal(1);
            txJson.lamportsPerSignature?.should.be.undefined();
            txJson.instructionsData.length.should.equal(8);
            txJson.instructionsData.should.deepEqual([
                {
                    type: 'NonceAdvance',
                    params: {
                        walletNonceAddress: '8Y7RM6JfcX4ASSNBkrkrmSbRu431YVi9Y3oLFnzC2dCh',
                        authWalletAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: 'CP5Dpaa42RtJmMuKqCQsLwma5Yh3knuvKsYDFX85F41S',
                        amount: '300000',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: '6B55XMiaS6tUZw5Tt3G1RaXAqdrvN38yXVDJmWvKLkiM',
                        amount: '300000',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: 'C1UjpxcXNBpp1UyvYsuNBNZ5Da1G1i49g3yTvC23Ny7e',
                        amount: '300000',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: 'CpUYXh9xXoWfkBVaBQRZ8nAgDbT16GZeQdqveeBS1hmk',
                        amount: '300000',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: '64s6NjmEokdhicHEd432X5Ut2EDfDmVqdvGh4rASn1gd',
                        amount: '300000',
                    },
                },
                {
                    type: 'Transfer',
                    params: {
                        fromAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        toAddress: '6nXxL2jMSdkgfHm13Twvn1gzRAPdrWnWLfu89PJL3Aqe',
                        amount: '300000',
                    },
                },
                {
                    type: 'Memo',
                    params: {
                        memo: 'test memo',
                    },
                },
            ]);
        });
    });
    describe('sign should', () => {
        it('fail if the tx doesnt have nonce', async () => {
            const tx = new lib_1.Transaction(coin);
            tx.solTransaction = new web3_js_1.Transaction();
            tx.solTransaction.feePayer = new web3_js_1.PublicKey(testData.authAccount.pub);
            const kp = new lib_1.KeyPair({ prv: testData.prvKeys.prvKey1.base58 });
            await tx.sign(kp).should.be.rejectedWith('Nonce is required before signing');
        });
        it('fail if the tx doesnt have feePayer', async () => {
            const tx = new lib_1.Transaction(coin);
            const kp = new lib_1.KeyPair({ prv: testData.prvKeys.prvKey1.base58 });
            tx.solTransaction = new web3_js_1.Transaction();
            tx.solTransaction.recentBlockhash = testData.blockHashes.validBlockHashes[0];
            await tx.sign(kp).should.be.rejectedWith('feePayer is required before signing');
        });
        it('fail if the KeyPair is not the right one', async () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_UNSIGNED);
            const keypair = new lib_1.KeyPair({ prv: testData.prvKeys.prvKey1.base58 });
            await tx.sign(keypair).should.be.rejectedWith('unknown signer: CP5Dpaa42RtJmMuKqCQsLwma5Yh3knuvKsYDFX85F41S');
        });
        it('fail if the KeyPair doesnt have a prv key', async () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_UNSIGNED);
            const keypair = new lib_1.KeyPair({ pub: testData.pubKeys.validPubKeys[0] });
            await tx.sign(keypair).should.be.rejectedWith('Missing private key');
        });
        it('succeed to sign with 1 KeyPair', async () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_UNSIGNED);
            const keypair = new lib_1.KeyPair({ prv: testData.accountWithSeed.privateKey.base58 });
            await tx.sign(keypair).should.be.fulfilled();
            should_1.default.equal(tx.toBroadcastFormat(), testData.RAW_TX_SIGNED);
        });
        it('succeed when try to sign with the same keyPair multiple times ', async () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_UNSIGNED);
            const keypair = new lib_1.KeyPair({ prv: testData.accountWithSeed.privateKey.base58 });
            await tx.sign([keypair, keypair, keypair, keypair]).should.be.fulfilled();
            should_1.default.equal(tx.toBroadcastFormat(), testData.RAW_TX_SIGNED);
        });
        it('succeed when try to sign with a keyPair that already signed', async () => {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_SIGNED);
            const keypair = new lib_1.KeyPair({ prv: testData.accountWithSeed.privateKey.base58 });
            await tx.sign(keypair).should.be.fulfilled();
            should_1.default.equal(tx.toBroadcastFormat(), testData.RAW_TX_SIGNED);
        });
    });
    describe('transaction parsing', function () {
        it('fromRawTransaction and toBroadcastFormat', async function () {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_UNSIGNED);
            should_1.default.equal(tx.toBroadcastFormat(), testData.RAW_TX_UNSIGNED);
        });
        it('fromRawTransaction, sign and toBroadcastFormat ', async function () {
            const tx = new lib_1.Transaction(coin);
            tx.fromRawTransaction(testData.RAW_TX_UNSIGNED);
            const keypair = new lib_1.KeyPair({ prv: testData.accountWithSeed.privateKey.base58 });
            await tx.sign(keypair);
            should_1.default.equal(tx.toBroadcastFormat(), testData.RAW_TX_SIGNED);
        });
    });
    describe('explain transaction', function () {
        const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsol');
        const blockHash = testData.blockHashes.validBlockHashes[0];
        const sender = testData.authAccount.pub;
        const address = testData.addresses.validAddresses[0];
        const amount = '10000';
        const wallet = new lib_1.KeyPair(testData.authAccount).getKeys();
        const stakeAccount = new lib_1.KeyPair(testData.stakeAccount).getKeys();
        const validator = testData.validator;
        it('should explain single transfer transaction', async function () {
            const tx = await factory
                .getTransferBuilder()
                .nonce(blockHash)
                .sender(sender)
                .send({ address, amount })
                .fee({ amount: 5000 })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain single transfer with durable nonce transaction', async function () {
            const tx = await factory
                .getTransferBuilder()
                .nonce(blockHash, { walletNonceAddress: testData.nonceAccount.pub, authWalletAddress: sender })
                .sender(sender)
                .send({ address, amount })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '0',
                    feeRate: undefined,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: {
                    authWalletAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                    walletNonceAddress: '8Y7RM6JfcX4ASSNBkrkrmSbRu431YVi9Y3oLFnzC2dCh',
                },
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain multi transfer with durable nonce and memo transaction', async function () {
            const tx = await factory
                .getTransferBuilder()
                .nonce(blockHash, { walletNonceAddress: testData.nonceAccount.pub, authWalletAddress: sender })
                .sender(sender)
                .memo('memo text')
                .send({ address, amount })
                .send({ address: testData.addresses.validAddresses[1], amount })
                .send({ address: testData.addresses.validAddresses[2], amount })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '30000',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                    },
                    {
                        address: 'Azz9EmNuhtjoYrhWvidWx1Hfd14SNBsYyzXhA9Tnoca8',
                        amount: '10000',
                    },
                    {
                        address: '2n2xqWM9Z18LqxfJzkNrMMFWiDUFYA2k6WSgSnf6EnJs',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                    {
                        address: sender,
                        value: '10000',
                    },
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '0',
                    feeRate: undefined,
                },
                memo: 'memo text',
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: {
                    authWalletAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                    walletNonceAddress: '8Y7RM6JfcX4ASSNBkrkrmSbRu431YVi9Y3oLFnzC2dCh',
                },
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain signed transfer transaction', async function () {
            const tx = await factory
                .getTransferBuilder()
                .fee({ amount: 5000 })
                .nonce(blockHash)
                .sender(sender)
                .send({ address, amount })
                .build();
            await tx.sign(new lib_1.KeyPair({ prv: testData.authAccount.prv }));
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: tx.id,
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain wallet init transaction', async function () {
            const tx = await factory
                .getWalletInitializationBuilder()
                .fee({ amount: 5000 })
                .sender(sender)
                .nonce(blockHash)
                .address(testData.addresses.validAddresses[1])
                .amount(amount)
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'WalletInitialization',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: 'Azz9EmNuhtjoYrhWvidWx1Hfd14SNBsYyzXhA9Tnoca8',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '10000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain wallet init with durable nonce and memo transaction', async function () {
            const tx = await factory
                .getWalletInitializationBuilder()
                .sender(sender)
                .nonce(blockHash, { walletNonceAddress: testData.nonceAccount.pub, authWalletAddress: sender })
                .memo('memo text')
                .address(testData.addresses.validAddresses[1])
                .amount(amount)
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'WalletInitialization',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: 'Azz9EmNuhtjoYrhWvidWx1Hfd14SNBsYyzXhA9Tnoca8',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '0',
                    feeRate: undefined,
                },
                memo: 'memo text',
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: {
                    authWalletAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                    walletNonceAddress: '8Y7RM6JfcX4ASSNBkrkrmSbRu431YVi9Y3oLFnzC2dCh',
                },
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain signed wallet init transaction', async function () {
            const tx = await factory
                .getWalletInitializationBuilder()
                .fee({ amount: 5000 })
                .sender(sender)
                .nonce(blockHash)
                .address(testData.addresses.validAddresses[1])
                .amount(amount)
                .build();
            await tx.sign(new lib_1.KeyPair({ prv: testData.authAccount.prv }));
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: tx.id,
                type: 'WalletInitialization',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: 'Azz9EmNuhtjoYrhWvidWx1Hfd14SNBsYyzXhA9Tnoca8',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '10000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain create ATA transaction with tokenEnablements field', async function () {
            const tx = await factory
                .getAtaInitializationBuilder()
                .fee({ amount: 5000 })
                .sender(sender)
                .nonce(blockHash)
                .mint('tsol:usdc')
                .rentExemptAmount(amount)
                .build();
            const explainedTransaction = tx.explainTransaction();
            // Verify the tokenEnablements field exists and contains the correct data
            explainedTransaction.should.have.property('tokenEnablements');
            explainedTransaction.tokenEnablements.should.be.an.Array();
            explainedTransaction.tokenEnablements.length.should.equal(1);
            explainedTransaction.tokenEnablements[0].should.have.properties(['address', 'tokenName', 'tokenAddress']);
            explainedTransaction.tokenEnablements[0].tokenName.should.equal('tsol:usdc');
            // Verify displayOrder includes tokenEnablements
            explainedTransaction.displayOrder.should.containEql('tokenEnablements');
            // Verify other important fields
            explainedTransaction.id.should.equal('UNAVAILABLE');
            explainedTransaction.type.should.equal('AssociatedTokenAccountInitialization');
            explainedTransaction.outputs.should.be.an.Array().and.be.empty();
            explainedTransaction.fee.fee.should.equal('15000');
            explainedTransaction.fee.feeRate.should.equal(5000);
            explainedTransaction.blockhash.should.equal('5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen');
            should_1.default.not.exist(explainedTransaction.durableNonce);
        });
        it('should explain multiple token enablements in a transaction', async function () {
            // Create a transaction with multiple token enablements
            const txBuilder = factory.getAtaInitializationBuilder();
            txBuilder.nonce(blockHash);
            txBuilder.sender(sender);
            txBuilder.enableToken({
                ownerAddress: sender,
                tokenName: 'tsol:usdc',
                ataAddress: 'DfGJHsrY7RchVwPfigT4T3Wkc4zGq1kkbWzNaVQrDE4w',
            });
            txBuilder.enableToken({
                ownerAddress: sender,
                tokenName: 'tsol:ray',
                ataAddress: 'GHhhPAuh5PdyeCM4zHPT7jz3jAGrXTPQu6mUMTtZ8TTf',
            });
            txBuilder.rentExemptAmount(amount);
            txBuilder.fee({ amount: 5000 });
            const tx = await txBuilder.build();
            const explainedTransaction = tx.explainTransaction();
            // Verify the tokenEnablements field contains multiple entries
            explainedTransaction.should.have.property('tokenEnablements');
            explainedTransaction.tokenEnablements.should.be.an.Array();
            explainedTransaction.tokenEnablements.length.should.equal(2);
            // Check the first token enablement
            explainedTransaction.tokenEnablements[0].should.have.properties(['address', 'tokenName', 'tokenAddress']);
            explainedTransaction.tokenEnablements[0].tokenName.should.equal('tsol:usdc');
            // Check the second token enablement
            explainedTransaction.tokenEnablements[1].should.have.properties(['address', 'tokenName', 'tokenAddress']);
            explainedTransaction.tokenEnablements[1].tokenName.should.equal('tsol:ray');
            // Verify outputs is empty (since token enablements are no longer added to outputs)
            explainedTransaction.outputs.should.be.an.Array();
            explainedTransaction.outputs.should.be.empty();
            // Display order should include tokenEnablements
            explainedTransaction.displayOrder.should.containEql('tokenEnablements');
        });
        it('should activate builder ', async function () {
            const tx = await factory
                .getStakingActivateBuilder()
                .stakingAddress(stakeAccount.pub)
                .sender(wallet.pub)
                .nonce(blockHash)
                .amount(amount)
                .validator(validator.pub)
                .fee({ amount: 5000 })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'StakingActivate',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: '7dRuGFbU2y2kijP6o1LYNzVyz4yf13MooqoionCzv5Za',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '10000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should deactivate builder ', async function () {
            const recentBlockHash = 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi';
            const tx = await factory
                .getStakingDeactivateBuilder()
                .stakingAddress(stakeAccount.pub)
                .sender(wallet.pub)
                .stakingAddress(stakeAccount.pub)
                .nonce(recentBlockHash)
                .fee({ amount: 5000 })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'StakingDeactivate',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '0',
                outputs: [],
                inputs: [],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain withdraw transaction ', async function () {
            const recentBlockHash = 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi';
            const tx = await factory
                .getStakingWithdrawBuilder()
                .stakingAddress(stakeAccount.pub)
                .sender(wallet.pub)
                .amount(amount)
                .nonce(recentBlockHash)
                .fee({ amount: 5000 })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'StakingWithdraw',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: stakeAccount.pub,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain withdraw transaction with memo and durable nonce ', async function () {
            const tx = await factory
                .getStakingWithdrawBuilder()
                .stakingAddress(stakeAccount.pub)
                .sender(wallet.pub)
                .amount(amount)
                .nonce(blockHash, { walletNonceAddress: testData.nonceAccount.pub, authWalletAddress: sender })
                .memo('memo text')
                .fee({ amount: 5000 })
                .build();
            tx.signablePayload;
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'StakingWithdraw',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '10000',
                outputs: [
                    {
                        address: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                        amount: '10000',
                    },
                ],
                inputs: [
                    {
                        address: stakeAccount.pub,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: 'memo text',
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: {
                    walletNonceAddress: testData.nonceAccount.pub,
                    authWalletAddress: sender,
                },
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain single token transfer transaction', async function () {
            const tx = await factory
                .getTokenTransferBuilder()
                .nonce(blockHash)
                .sender(sender)
                .send({ address, amount, tokenName: 'tsol:usdc' })
                .fee({ amount: 5000 })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '0',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                        tokenName: 'tsol:usdc',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain single token transfer transaction with optional params', async function () {
            const tx = await factory
                .getTokenTransferBuilder()
                .nonce(blockHash)
                .sender(sender)
                .send({
                address,
                amount,
                tokenName: 'tsol:ams',
                tokenAddress: 'F4uLeXioFz3hw13MposuwaQbMcZbCjqvEGPPeRRB1Byf',
                programId: 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
                decimalPlaces: 9,
            })
                .fee({ amount: 5000 })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '0',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                        // WASM path resolves token name via @bitgo/statics; this mint is not registered,
                        // so the raw mint address is returned instead of the human-readable name.
                        tokenName: 'F4uLeXioFz3hw13MposuwaQbMcZbCjqvEGPPeRRB1Byf',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '5000',
                    feeRate: 5000,
                },
                memo: undefined,
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: undefined,
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
        it('should explain multi token transfer with durable nonce and memo transaction', async function () {
            const tx = await factory
                .getTokenTransferBuilder()
                .nonce(blockHash, { walletNonceAddress: testData.nonceAccount.pub, authWalletAddress: sender })
                .sender(sender)
                .memo('memo text')
                .send({ address, amount, tokenName: 'tsol:usdc' })
                .send({ address: testData.addresses.validAddresses[1], amount, tokenName: 'tsol:usdc' })
                .send({ address: testData.addresses.validAddresses[2], amount, tokenName: 'tsol:usdc' })
                .build();
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: [
                    'id',
                    'type',
                    'blockhash',
                    'durableNonce',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'tokenEnablements',
                    'fee',
                    'memo',
                ],
                id: 'UNAVAILABLE',
                type: 'Send',
                changeOutputs: [],
                changeAmount: '0',
                outputAmount: '0',
                outputs: [
                    {
                        address: 'DesU7XscZjng8yj5VX6AZsk3hWSW4sQ3rTG2LuyQ2P4H',
                        amount: '10000',
                        tokenName: 'tsol:usdc',
                    },
                    {
                        address: 'Azz9EmNuhtjoYrhWvidWx1Hfd14SNBsYyzXhA9Tnoca8',
                        amount: '10000',
                        tokenName: 'tsol:usdc',
                    },
                    {
                        address: '2n2xqWM9Z18LqxfJzkNrMMFWiDUFYA2k6WSgSnf6EnJs',
                        amount: '10000',
                        tokenName: 'tsol:usdc',
                    },
                ],
                inputs: [
                    {
                        address: sender,
                        value: '10000',
                    },
                    {
                        address: sender,
                        value: '10000',
                    },
                    {
                        address: sender,
                        value: '10000',
                    },
                ],
                feePayer: sender,
                fee: {
                    fee: '0',
                    feeRate: undefined,
                },
                memo: 'memo text',
                blockhash: '5ne7phA48Jrvpn39AtupB8ZkCCAy8gLTfpGihZPuDqen',
                durableNonce: {
                    authWalletAddress: '5hr5fisPi6DXNuuRpm5XUbzpiEnmdyxXuBDTwzwZj5Pe',
                    walletNonceAddress: '8Y7RM6JfcX4ASSNBkrkrmSbRu431YVi9Y3oLFnzC2dCh',
                },
                tokenEnablements: [],
                ataOwnerMap: {},
            });
        });
    });
});
