// @bitgo/sdk-coin-sol/dist/cjs/test/unit/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const lib_1 = require("../../src/lib");
const testData = __importStar(require("../resources/sol"));
const web3_js_1 = require("@solana/web3.js");
const constants_1 = require("../../src/lib/constants");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const spl_token_1 = require("@solana/spl-token");
describe('SOL util library', function () {
    describe('isValidAddress', function () {
        it('should fail to validate invalid addresses', function () {
            for (const address of testData.addresses.invalidAddresses) {
                should_1.default.doesNotThrow(() => lib_1.Utils.isValidAddress(address));
                should_1.default.equal(lib_1.Utils.isValidAddress(address), false);
            }
            // @ts-expect-error Testing for missing param, should not throw an error
            should_1.default.doesNotThrow(() => lib_1.Utils.isValidAddress(undefined));
            // @ts-expect-error Testing for missing param, should return false
            should_1.default.equal(lib_1.Utils.isValidAddress(undefined), false);
        });
        it('should succeed to validate valid addresses', function () {
            for (const address of testData.addresses.validAddresses) {
                should_1.default.equal(lib_1.Utils.isValidAddress(address), true);
            }
        });
    });
    describe('isValidBlockId', function () {
        it('should fail to validate invalid Block hashes', function () {
            for (const blockHash of testData.blockHashes.invalidBlockHashes) {
                should_1.default.doesNotThrow(() => lib_1.Utils.isValidBlockId(blockHash));
                should_1.default.equal(lib_1.Utils.isValidBlockId(blockHash), false);
            }
        });
        it('should succeed to validate valid Block hashes', function () {
            for (const blockHash of testData.blockHashes.validBlockHashes) {
                should_1.default.equal(lib_1.Utils.isValidBlockId(blockHash), true);
            }
        });
    });
    describe('isValidPublicKey', function () {
        it('should fail to validate invalid public keys', function () {
            for (const pubKey of testData.pubKeys.invalidPubKeys) {
                should_1.default.doesNotThrow(() => lib_1.Utils.isValidPublicKey(pubKey));
                should_1.default.equal(lib_1.Utils.isValidPublicKey(pubKey), false);
            }
            // @ts-expect-error Testing for missing param, should not throw an error
            should_1.default.doesNotThrow(() => lib_1.Utils.isValidPublicKey(undefined));
            // @ts-expect-error Testing for missing param, should return false
            should_1.default.equal(lib_1.Utils.isValidPublicKey(undefined), false);
        });
        it('should succeed to validate public keys', function () {
            for (const pubKey of testData.pubKeys.validPubKeys) {
                should_1.default.equal(lib_1.Utils.isValidPublicKey(pubKey), true);
            }
        });
    });
    describe('isValidPrivateKey', function () {
        it('should fail to validate invalid private keys', function () {
            for (const prvKey of testData.prvKeys.invalidPrvKeys) {
                should_1.default.doesNotThrow(() => lib_1.Utils.isValidPrivateKey(prvKey));
                should_1.default.equal(lib_1.Utils.isValidPrivateKey(prvKey), false);
            }
        });
        it('should succeed to validate private keys', function () {
            const validPrvKey = [testData.prvKeys.prvKey1.base58, testData.prvKeys.prvKey1.uint8Array];
            for (const prvKey of validPrvKey) {
                should_1.default.equal(lib_1.Utils.isValidPrivateKey(prvKey), true);
            }
        });
    });
    describe('isValidRawTransaction', function () {
        it('should fail to validate an invalid raw transaction', function () {
            should_1.default.doesNotThrow(() => lib_1.Utils.isValidRawTransaction(testData.INVALID_RAW_TX));
            should_1.default.equal(lib_1.Utils.isValidRawTransaction(testData.INVALID_RAW_TX), false);
        });
        it('should succeed to validate a valid raw transaction', function () {
            const validRawTxs = [testData.RAW_TX_UNSIGNED, testData.RAW_TX_SIGNED];
            for (const rawTx of validRawTxs) {
                should_1.default.equal(lib_1.Utils.isValidRawTransaction(rawTx), true);
            }
        });
    });
    describe('isValidSignature and isValidTransactionId', function () {
        it('should fail to validate invalid signatures', function () {
            for (const signature of testData.signatures.invalidSignatures) {
                should_1.default.doesNotThrow(() => lib_1.Utils.isValidSignature(signature));
                should_1.default.equal(lib_1.Utils.isValidSignature(signature), false);
                should_1.default.doesNotThrow(() => lib_1.Utils.isValidTransactionId(signature));
                should_1.default.equal(lib_1.Utils.isValidTransactionId(signature), false);
            }
        });
        it('should succeed to validate valid signatures', function () {
            for (const signature of testData.signatures.validSignatures) {
                should_1.default.equal(lib_1.Utils.isValidSignature(signature), true);
                should_1.default.equal(lib_1.Utils.isValidTransactionId(signature), true);
            }
        });
    });
    describe('base58 and Uint8Array encoding', function () {
        it('should succeed to base58ToUint8Array', function () {
            should_1.default.deepEqual(lib_1.Utils.base58ToUint8Array(testData.prvKeys.prvKey1.base58), testData.prvKeys.prvKey1.uint8Array);
        });
        it('should succeed to Uint8ArrayTobase58', function () {
            should_1.default.deepEqual(lib_1.Utils.Uint8ArrayTobase58(testData.prvKeys.prvKey1.uint8Array), testData.prvKeys.prvKey1.base58);
        });
    });
    describe('isValidAmount', function () {
        it('should succeed for valid amounts', function () {
            const validAmounts = ['0', '12312312'];
            for (const amount of validAmounts) {
                should_1.default.equal(lib_1.Utils.isValidAmount(amount), true);
            }
        });
        it('should fail for invalid amounts', function () {
            const invalidAmounts = ['-1', 'randomstring', '33.04235'];
            for (const amount of invalidAmounts) {
                should_1.default.equal(lib_1.Utils.isValidAmount(amount), false);
            }
        });
    });
    describe('verifySignature', function () {
        it('should succeed for valid signature in a unsigned tx', function () {
            const signature = '335sxAuVj5ucXqVWW82QwpFLArPbdD3gXfXr4KrxkLkUpmLB3Nwz2G82z2TqiDD7mNAAbHkcAqD5ycDZp1vVKtjf';
            lib_1.Utils.verifySignature(testData.TOKEN_TRANSFER_UNSIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE_OLD, signature, testData.associatedTokenAccounts.accounts[0].pub).should.equal(true);
        });
        it('should succeed for valid signature in a signed tx', function () {
            const signature = '335sxAuVj5ucXqVWW82QwpFLArPbdD3gXfXr4KrxkLkUpmLB3Nwz2G82z2TqiDD7mNAAbHkcAqD5ycDZp1vVKtjf';
            lib_1.Utils.verifySignature(testData.TOKEN_TRANSFER_SIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE_OLD, signature, testData.associatedTokenAccounts.accounts[0].pub).should.equal(true);
        });
        it('should fail for invalid signature', function () {
            const signature = '2QdKALq4adaTahJH13AGzM5bAFuNshw43iQBdVS9D2Loq736zUgPXfHj32cNJKX6FyjUzYJhGfEyAAB5FgYUW6zR';
            lib_1.Utils.verifySignature(testData.TRANSFER_UNSIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE, signature, testData.authAccount.pub).should.equal(false);
        });
        it('should fail for invalid pub key', function () {
            const signature = '3pD6ayWtvFkn8Fe5efYbSaCYMpnDwzDTmmeoMhcSMAcMrGvmwPFhLxok5vxhHnooA3YSXfnyZgi4e3K3sCHmgU3k';
            lib_1.Utils.verifySignature(testData.TRANSFER_UNSIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE, signature, testData.nonceAccount.pub).should.equal(false);
        });
        it('should throw for invalid tx', function () {
            const signature = '3pD6ayWtvFkn8Fe5efYbSaCYMpnDwzDTmmeoMhcSMAcMrGvmwPFhLxok5vxhHnooA3YSXfnyZgi4e3K3sCHmgU3k';
            (0, should_1.default)(() => lib_1.Utils.verifySignature(testData.INVALID_RAW_TX, signature, testData.nonceAccount.pub)).throwError('Invalid serializedTx');
        });
        it('should throw for invalid pubkey', function () {
            const signature = '3pD6ayWtvFkn8Fe5efYbSaCYMpnDwzDTmmeoMhcSMAcMrGvmwPFhLxok5vxhHnooA3YSXfnyZgi4e3K3sCHmgU3k';
            (0, should_1.default)(() => lib_1.Utils.verifySignature(testData.TRANSFER_UNSIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE, signature, testData.pubKeys.invalidPubKeys[0])).throwError('Invalid publicKey');
        });
        it('should throw for invalid signature', function () {
            const signature = 'randomstring';
            (0, should_1.default)(() => lib_1.Utils.verifySignature(testData.TRANSFER_UNSIGNED_TX_WITH_MEMO_AND_DURABLE_NONCE, signature, testData.nonceAccount.pub)).throwError('Invalid signature');
        });
    });
    describe('isValidMemo', function () {
        it('should return true for valid memo', function () {
            lib_1.Utils.isValidMemo('testmemo').should.equal(true);
        });
        it('should return false for a long memo', function () {
            lib_1.Utils.isValidMemo('3pD6ayWtvFkn8Fe5efYbSaCYMpnDwzDTmmeoMhcSMAcMrGvmwPFhLxok5vxhHnooA3YSXfnyZgi4e3K3sCHmgU3kPFhLxok5vxhHnooA3YSXfnyZgi4e3K3sCHmgU3k').should.equal(true);
        });
    });
    describe('getInstructionType', function () {
        it('should succeed for memo program', function () {
            const memoInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                data: Buffer.from('random memo'),
            });
            lib_1.Utils.getInstructionType(memoInstruction).should.equal('Memo');
        });
        it('should succeed for system program', function () {
            const fromAddress = testData.authAccount.pub;
            const toAddress = testData.nonceAccount.pub;
            const amount = '100000';
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(fromAddress),
                toPubkey: new web3_js_1.PublicKey(toAddress),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            });
            lib_1.Utils.getInstructionType(transferInstruction).should.equal('Transfer');
        });
        it('should fallback to customInstruction for unsupported instructionType', function () {
            const voteAddress = 'Vote111111111111111111111111111111111111111';
            const invalidInstruction = new web3_js_1.TransactionInstruction({
                keys: [],
                programId: new web3_js_1.PublicKey(voteAddress),
                data: Buffer.from('random memo'),
            });
            lib_1.Utils.getInstructionType(invalidInstruction).should.equal('CustomInstruction');
        });
        it('should succeed for ComputeBudget SetComputeUnitLimit instruction', function () {
            const setComputeUnitLimitInstruction = web3_js_1.ComputeBudgetProgram.setComputeUnitLimit({
                units: 300000,
            });
            lib_1.Utils.getInstructionType(setComputeUnitLimitInstruction).should.equal('SetComputeUnitLimit');
        });
        it('should succeed for ComputeBudget SetComputeUnitPrice instruction', function () {
            const setComputeUnitPriceInstruction = web3_js_1.ComputeBudgetProgram.setComputeUnitPrice({
                microLamports: 100000,
            });
            lib_1.Utils.getInstructionType(setComputeUnitPriceInstruction).should.equal('SetPriorityFee');
        });
    });
    describe('validateIntructionTypes', function () {
        it('should succeed for valid instruction type', function () {
            const fromAddress = testData.authAccount.pub;
            const toAddress = testData.nonceAccount.pub;
            const amount = '100000';
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(fromAddress),
                toPubkey: new web3_js_1.PublicKey(toAddress),
                lamports: new bignumber_js_1.default(amount).toNumber(),
            });
            should_1.default.doesNotThrow(() => lib_1.Utils.validateIntructionTypes([transferInstruction]));
        });
        it('should fail for invalid instruction type', function () {
            const accountPubkey = testData.authAccount.pub;
            const assignInstruction = web3_js_1.SystemProgram.nonceWithdraw({
                noncePubkey: new web3_js_1.PublicKey(accountPubkey),
                authorizedPubkey: new web3_js_1.PublicKey(accountPubkey),
                toPubkey: new web3_js_1.PublicKey(accountPubkey),
                lamports: 200,
            });
            (0, should_1.default)(() => lib_1.Utils.validateIntructionTypes([assignInstruction])).throwError('Invalid transaction, instruction type not supported: ' + lib_1.Utils.getInstructionType(assignInstruction));
        });
    });
    describe('validateRawTransaction', function () {
        it('should succeed for valid raw transaction', function () {
            should_1.default.doesNotThrow(() => lib_1.Utils.validateRawTransaction(testData.RAW_TX_UNSIGNED));
        });
        it('should fail for invalid raw transaction', function () {
            (0, should_1.default)(() => lib_1.Utils.validateRawTransaction('AAAAAAAAAAAAAAAAA')).throwError('Invalid raw transaction');
        });
        it('should fail for missing param', function () {
            // @ts-expect-error Testing for missing param, should throw error
            (0, should_1.default)(() => lib_1.Utils.validateRawTransaction()).throwError('Invalid raw transaction: Undefined');
        });
    });
    describe('getSolTokenFromTokenName', function () {
        it('should succeed for sol token', function () {
            should_1.default.notEqual(lib_1.Utils.getSolTokenFromTokenName('tsol:usdc'), undefined);
        });
        it('should fail for non tokens', function () {
            should_1.default.equal(lib_1.Utils.getSolTokenFromTokenName('tsol'), undefined);
        });
        it('should fail if tokenName is not in coins', function () {
            should_1.default.equal(lib_1.Utils.getSolTokenFromTokenName('something random'), undefined);
        });
    });
    describe('getAssociatedTokenAccountAddress', function () {
        const usdcMintAddress = testData.tokenTransfers.mintUSDC;
        const tokenAddress = '141BFNem3pknc8CzPVLv1Ri3btgKdCsafYP5nXwmXfxU';
        it('should succeed for native address as owner address', async function () {
            const ownerAddress = testData.authAccount.pub;
            const result = await lib_1.Utils.getAssociatedTokenAccountAddress(usdcMintAddress, ownerAddress);
            result.should.be.equal(tokenAddress);
        });
        it('should fail for token address as owner address', async function () {
            const invalidOwnerAddress = tokenAddress;
            await lib_1.Utils.getAssociatedTokenAccountAddress(usdcMintAddress, invalidOwnerAddress).should.be.rejectedWith('Invalid ownerAddress - address off ed25519 curve, got: ' + invalidOwnerAddress);
        });
    });
    describe('getAssociatedTokenAccountAddress for sol 2022 token', function () {
        const ai16ZMintAddress = testData.tokenTransfers.mintAI16Z;
        const tokenAddress = 'JC2HAixRNrhBgNjCsV2PBfZEdyitEnWqJPNGokFD8xSA';
        it('should succeed for native address as owner address', async function () {
            const ownerAddress = testData.sol2022AuthAccount.pub;
            const result = await lib_1.Utils.getAssociatedTokenAccountAddress(ai16ZMintAddress, ownerAddress, false, spl_token_1.TOKEN_2022_PROGRAM_ID.toString());
            result.should.be.equal(tokenAddress);
        });
        it('should fail for token address as owner address', async function () {
            const invalidOwnerAddress = tokenAddress;
            await lib_1.Utils.getAssociatedTokenAccountAddress(ai16ZMintAddress, invalidOwnerAddress, false, spl_token_1.TOKEN_2022_PROGRAM_ID.toString()).should.be.rejectedWith('Invalid ownerAddress - address off ed25519 curve, got: ' + invalidOwnerAddress);
        });
    });
    describe('getAssociatedTokenAccountAddress for sol 2022 token off curve address', function () {
        const mintAddress = testData.sol2022PumpTransfers.mint;
        const tokenAddress = '6fxrV8zFVo7Q7U9RzgfNiZGHFygdCzXaFZHyiFhNZjNF';
        it('should succeed for native address as owner address', async function () {
            const ownerAddress = testData.sol2022PumpTransfers.owner;
            const result = await lib_1.Utils.getAssociatedTokenAccountAddress(mintAddress, ownerAddress, true, spl_token_1.TOKEN_2022_PROGRAM_ID.toString());
            result.should.be.equal(tokenAddress);
        });
    });
    describe('matchTransactionTypeByInstructionsOrder', function () {
        describe('Activate stake instructions', function () {
            it('should match staking activate instructions', function () {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                const validator = new web3_js_1.PublicKey(testData.validator.pub);
                const amount = '100000';
                // Instructions
                const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                    fromPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                    authorized: {
                        staker: fromAccount,
                        withdrawer: fromAccount,
                    },
                    lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                    lamports: new bignumber_js_1.default(amount).toNumber(),
                }).instructions;
                const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                    authorizedPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                    votePubkey: validator,
                }).instructions;
                const instructions = [...stakingActivateInstructions, ...stakingDelegateInstructions];
                const isAMatch = lib_1.Utils.matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingActivateInstructionsIndexes);
                isAMatch.should.be.true();
            });
            it('should match staking activate instructions with memo and durable nonce', function () {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const nonceAccount = testData.nonceAccount.pub;
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                const validator = new web3_js_1.PublicKey(testData.validator.pub);
                const amount = '100000';
                const memo = 'test memo';
                // Instructions
                const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                    noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                    authorizedPubkey: fromAccount,
                });
                const stakingActivateInstructions = web3_js_1.StakeProgram.createAccount({
                    fromPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                    authorized: {
                        staker: fromAccount,
                        withdrawer: fromAccount,
                    },
                    lockup: new web3_js_1.Lockup(0, 0, fromAccount),
                    lamports: new bignumber_js_1.default(amount).toNumber(),
                }).instructions;
                const stakingDelegateInstructions = web3_js_1.StakeProgram.delegate({
                    authorizedPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                    votePubkey: validator,
                }).instructions;
                const memoInstruction = new web3_js_1.TransactionInstruction({
                    keys: [],
                    programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                    data: Buffer.from(memo),
                });
                const instructions = [
                    nonceAdvanceInstruction,
                    ...stakingActivateInstructions,
                    ...stakingDelegateInstructions,
                    memoInstruction,
                ];
                const isAMatch = lib_1.Utils.matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingActivateInstructionsIndexes);
                isAMatch.should.be.true();
            });
        });
        describe('Deactivate stake instructions', function () {
            it('should match staking deactivate instructions', function () {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                // Instructions
                const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                    authorizedPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                }).instructions;
                const instructions = [...stakingDeactivateInstructions];
                const isAMatch = lib_1.Utils.matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingDeactivateInstructionsIndexes);
                isAMatch.should.be.true();
            });
            it('should match staking deactivate instructions with memo and durable nonce', function () {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const nonceAccount = testData.nonceAccount.pub;
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                const memo = 'test memo';
                // Instructions
                const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                    noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                    authorizedPubkey: fromAccount,
                });
                const stakingDeactivateInstructions = web3_js_1.StakeProgram.deactivate({
                    authorizedPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                }).instructions;
                const memoInstruction = new web3_js_1.TransactionInstruction({
                    keys: [],
                    programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                    data: Buffer.from(memo),
                });
                const instructions = [nonceAdvanceInstruction, ...stakingDeactivateInstructions, memoInstruction];
                const isAMatch = lib_1.Utils.matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingDeactivateInstructionsIndexes);
                isAMatch.should.be.true();
            });
        });
        describe('Staking withdraw instructions', function () {
            it('should match staking withdraw instructions', function () {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                const amount = '100000';
                // Instructions
                const stakingWithdrawInstructions = web3_js_1.StakeProgram.withdraw({
                    authorizedPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                    toPubkey: fromAccount,
                    lamports: new bignumber_js_1.default(amount).toNumber(),
                }).instructions;
                const instructions = [...stakingWithdrawInstructions];
                const isAMatch = lib_1.Utils.matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingWithdrawInstructionsIndexes);
                isAMatch.should.be.true();
            });
            it('should match staking withdraw instructions with memo and durable nonce', function () {
                const fromAccount = new web3_js_1.PublicKey(testData.authAccount.pub);
                const nonceAccount = testData.nonceAccount.pub;
                const stakingAccount = new web3_js_1.PublicKey(testData.stakeAccount.pub);
                const amount = '100000';
                const memo = 'test memo';
                // Instructions
                const nonceAdvanceInstruction = web3_js_1.SystemProgram.nonceAdvance({
                    noncePubkey: new web3_js_1.PublicKey(nonceAccount),
                    authorizedPubkey: fromAccount,
                });
                // Instructions
                const stakingWithdrawInstructions = web3_js_1.StakeProgram.withdraw({
                    authorizedPubkey: fromAccount,
                    stakePubkey: stakingAccount,
                    toPubkey: fromAccount,
                    lamports: new bignumber_js_1.default(amount).toNumber(),
                }).instructions;
                const memoInstruction = new web3_js_1.TransactionInstruction({
                    keys: [],
                    programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
                    data: Buffer.from(memo),
                });
                const instructions = [nonceAdvanceInstruction, ...stakingWithdrawInstructions, memoInstruction];
                const isAMatch = lib_1.Utils.matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingWithdrawInstructionsIndexes);
                isAMatch.should.be.true();
            });
        });
    });
    describe('isIdempotentAtaInstruction', function () {
        const mockKeys = [
            { pubkey: new web3_js_1.PublicKey('9i8CSiz2un7rfuNvTMt1tTXbHSaMkPZyJ4MexY1yeZBD'), isSigner: true, isWritable: true },
            { pubkey: new web3_js_1.PublicKey('3F7X7ifwMR29Z3t1YamFg6yzCcsSkjAZpZF8yU1kWURh'), isSigner: false, isWritable: true },
        ];
        it('should return true for idempotent ATA instruction with discriminator byte 1', function () {
            const instruction = new web3_js_1.TransactionInstruction({
                programId: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                keys: mockKeys,
                data: Buffer.from([1]),
            });
            lib_1.Utils.isIdempotentAtaInstruction(instruction).should.equal(true);
        });
        it('should return true for idempotent ATA instruction from base64 "AQ=="', function () {
            const instruction = new web3_js_1.TransactionInstruction({
                programId: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                keys: mockKeys,
                data: Buffer.from('AQ==', 'base64'),
            });
            lib_1.Utils.isIdempotentAtaInstruction(instruction).should.equal(true);
        });
        it('should return false for legacy ATA instruction with empty data', function () {
            const instruction = new web3_js_1.TransactionInstruction({
                programId: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                keys: mockKeys,
                data: Buffer.from([]),
            });
            lib_1.Utils.isIdempotentAtaInstruction(instruction).should.equal(false);
        });
        it('should return false for instruction with wrong discriminator', function () {
            const instruction = new web3_js_1.TransactionInstruction({
                programId: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                keys: mockKeys,
                data: Buffer.from([2]),
            });
            lib_1.Utils.isIdempotentAtaInstruction(instruction).should.equal(false);
        });
        it('should return false for instruction with multiple bytes', function () {
            const instruction = new web3_js_1.TransactionInstruction({
                programId: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                keys: mockKeys,
                data: Buffer.from([1, 2]),
            });
            lib_1.Utils.isIdempotentAtaInstruction(instruction).should.equal(false);
        });
    });
});
