// @bitgo/sdk-coin-atom/dist/test/unit/transactionBuilder/transferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_core_1 = require("@bitgo/sdk-core");
const encoding_1 = require("@cosmjs/encoding");
const axios_1 = __importDefault(require("axios"));
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/atom"));
const getBuilderFactory_1 = require("../getBuilderFactory");
describe('Atom Transfer Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tatom');
    const testTx = testData.TEST_SEND_TX;
    const testTxWithMemo = testData.TEST_TX_WITH_MEMO;
    it('should build a Transfer tx with signature', async function () {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sequence(testTx.sequence);
        txBuilder.gasBudget(testTx.gasBudget);
        txBuilder.messages([testTx.sendMessage.value]);
        txBuilder.publicKey((0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.pubKey)));
        txBuilder.addSignature({ pub: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.pubKey)) }, Buffer.from(testTx.signature, 'base64'));
        const tx = await txBuilder.build();
        const json = await (await txBuilder.build()).toJson();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
        should_1.default.deepEqual(json.gasBudget, testTx.gasBudget);
        should_1.default.deepEqual(json.sendMessages, [testTx.sendMessage]);
        should_1.default.deepEqual(json.publicKey, (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.pubKey)));
        should_1.default.deepEqual(json.sequence, testTx.sequence);
        const rawTx = tx.toBroadcastFormat();
        should_1.default.equal(rawTx, testTx.signedTxBase64);
        should_1.default.deepEqual(tx.inputs, [
            {
                address: testData.TEST_SEND_TX.sender,
                value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
        should_1.default.deepEqual(tx.outputs, [
            {
                address: testData.TEST_SEND_TX.sendMessage.value.toAddress,
                value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
    });
    it('should build a Transfer tx with signature and memo', async function () {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sequence(testTxWithMemo.sequence);
        txBuilder.gasBudget(testTxWithMemo.gasBudget);
        txBuilder.messages([testTxWithMemo.sendMessage.value]);
        txBuilder.publicKey((0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTxWithMemo.pubKey)));
        txBuilder.memo(testTxWithMemo.memo);
        txBuilder.addSignature({ pub: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTxWithMemo.pubKey)) }, Buffer.from(testTxWithMemo.signature, 'base64'));
        const tx = await txBuilder.build();
        const json = await (await txBuilder.build()).toJson();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
        should_1.default.deepEqual(json.gasBudget, testTxWithMemo.gasBudget);
        should_1.default.deepEqual(json.sendMessages, [testTxWithMemo.sendMessage]);
        should_1.default.deepEqual(json.publicKey, (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTxWithMemo.pubKey)));
        should_1.default.deepEqual(json.sequence, testTxWithMemo.sequence);
        should_1.default.equal(json.memo, testTxWithMemo.memo);
        const rawTx = tx.toBroadcastFormat();
        should_1.default.equal(rawTx, testTxWithMemo.signedTxBase64);
        should_1.default.deepEqual(tx.inputs, [
            {
                address: testTxWithMemo.sendMessage.value.fromAddress,
                value: testTxWithMemo.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
        should_1.default.deepEqual(tx.outputs, [
            {
                address: testTxWithMemo.sendMessage.value.toAddress,
                value: testTxWithMemo.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
    });
    it('should build a Transfer tx without signature', async function () {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sequence(testTx.sequence);
        txBuilder.gasBudget(testTx.gasBudget);
        txBuilder.messages([testTx.sendMessage.value]);
        txBuilder.publicKey((0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.pubKey)));
        const tx = await txBuilder.build();
        const json = await (await txBuilder.build()).toJson();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
        should_1.default.deepEqual(json.gasBudget, testTx.gasBudget);
        should_1.default.deepEqual(json.sendMessages, [testTx.sendMessage]);
        should_1.default.deepEqual(json.publicKey, (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.pubKey)));
        should_1.default.deepEqual(json.sequence, testTx.sequence);
        tx.toBroadcastFormat();
        should_1.default.deepEqual(tx.inputs, [
            {
                address: testData.TEST_SEND_TX.sender,
                value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
        should_1.default.deepEqual(tx.outputs, [
            {
                address: testData.TEST_SEND_TX.sendMessage.value.toAddress,
                value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
    });
    it('should sign a Transfer tx', async function () {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sequence(testTx.sequence);
        txBuilder.gasBudget(testTx.gasBudget);
        txBuilder.messages([testTx.sendMessage.value]);
        txBuilder.accountNumber(testTx.accountNumber);
        txBuilder.chainId(testTx.chainId);
        txBuilder.sign({ key: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.privateKey)) });
        const tx = await txBuilder.build();
        const json = await (await txBuilder.build()).toJson();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
        should_1.default.deepEqual(json.gasBudget, testTx.gasBudget);
        should_1.default.deepEqual(json.sendMessages, [testTx.sendMessage]);
        should_1.default.deepEqual(json.publicKey, (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.pubKey)));
        should_1.default.deepEqual(json.sequence, testTx.sequence);
        const rawTx = tx.toBroadcastFormat();
        should_1.default.equal(tx.signature[0], (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.signature)));
        should_1.default.equal(rawTx, testTx.signedTxBase64);
        should_1.default.deepEqual(tx.inputs, [
            {
                address: testData.TEST_SEND_TX.sender,
                value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
        should_1.default.deepEqual(tx.outputs, [
            {
                address: testData.TEST_SEND_TX.sendMessage.value.toAddress,
                value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                coin: 'tatom',
            },
        ]);
    });
    it('should build a sendMany Transfer tx', async function () {
        const testSendManyTx = testData.TEST_SEND_MANY_TX;
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sequence(testSendManyTx.sequence);
        txBuilder.gasBudget(testSendManyTx.gasBudget);
        txBuilder.messages(testSendManyTx.sendMessages.map((msg) => msg.value));
        txBuilder.publicKey((0, encoding_1.toHex)((0, encoding_1.fromBase64)(testSendManyTx.pubKey)));
        txBuilder.chainId(testSendManyTx.chainId);
        txBuilder.accountNumber(testSendManyTx.accountNumber);
        txBuilder.memo(testSendManyTx.memo);
        txBuilder.addSignature({ pub: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testSendManyTx.pubKey)) }, Buffer.from(testSendManyTx.signature, 'base64'));
        const tx = await txBuilder.build();
        const json = await (await txBuilder.build()).toJson();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
        should_1.default.deepEqual(json.gasBudget, testSendManyTx.gasBudget);
        should_1.default.deepEqual(json.sendMessages, testSendManyTx.sendMessages);
        should_1.default.deepEqual(json.publicKey, (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testSendManyTx.pubKey)));
        should_1.default.deepEqual(json.sequence, testSendManyTx.sequence);
        should_1.default.deepEqual(tx.inputs, testSendManyTx.sendMessages.map((msg) => {
            return {
                address: msg.value.fromAddress,
                value: msg.value.amount[0].amount,
                coin: 'tatom',
            };
        }));
        should_1.default.deepEqual(tx.outputs, testSendManyTx.sendMessages.map((msg) => {
            return {
                address: msg.value.toAddress,
                value: msg.value.amount[0].amount,
                coin: 'tatom',
            };
        }));
        should_1.default.equal(tx.id, testSendManyTx.hash);
        const rawTx = tx.toBroadcastFormat();
        should_1.default.equal(rawTx, testSendManyTx.signedTxBase64);
    });
    xit('should submit a send transaction', async () => {
        const keyPair = new src_1.KeyPair({ prv: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(testTx.privateKey)) });
        const axiosConfig = {
            headers: {
                'Content-Type': 'application/cbor',
            },
            timeout: 10000,
        };
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sequence(9);
        txBuilder.gasBudget(testTx.gasBudget);
        txBuilder.messages([testTx.sendMessage.value]);
        txBuilder.accountNumber(testTx.accountNumber);
        txBuilder.chainId(testTx.chainId);
        txBuilder.sign({ key: keyPair.getKeys().prv });
        const tx = await txBuilder.build();
        const serializedTx = tx.toBroadcastFormat();
        try {
            const res = await axios_1.default.post('https://rest.sentry-01.theta-testnet.polypore.xyz/cosmos/tx/v1beta1/txs', {
                mode: 'BROADCAST_MODE_BLOCK',
                tx_bytes: serializedTx,
            }, axiosConfig);
            console.log(res);
        }
        catch (err) {
            console.log(err);
        }
    });
});
