// @bitgo/sdk-api/dist/src/api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toBitgoRequest = toBitgoRequest;
exports.handleResponseResult = handleResponseResult;
exports.handleResponseError = handleResponseError;
exports.serializeRequestData = serializeRequestData;
exports.setRequestQueryString = setRequestQueryString;
exports.verifyResponse = verifyResponse;
/**
 * @prettier
 */
const debug_1 = __importDefault(require("debug"));
const eol_1 = __importDefault(require("eol"));
const lodash_1 = __importDefault(require("lodash"));
const sanitize_html_1 = __importDefault(require("sanitize-html"));
const superagent_1 = __importDefault(require("superagent"));
const url_1 = __importDefault(require("url"));
const querystring_1 = __importDefault(require("querystring"));
const sdk_core_1 = require("@bitgo/sdk-core");
const debug = (0, debug_1.default)('bitgo:api');
/**
 * Add the bitgo-specific result() function on a superagent request.
 *
 * If the server response is successful, the `result()` function will return either the entire response body,
 * or the field from the response body specified by the `optionalField` parameter if it is provided.
 *
 * If the server response with an error, `result()` will handle HTTP errors appropriately by
 * rethrowing them as an `ApiResponseError` if possible, and otherwise rethrowing the underlying response error.
 *
 * @param req
 */
function toBitgoRequest(req) {
    return Object.assign(req, {
        result(optionalField) {
            return req.then((response) => handleResponseResult(optionalField)(response), (error) => handleResponseError(error));
        },
    });
}
/**
 * Return a function which extracts the specified response body property from the response if successful,
 * otherwise throw an `ApiErrorResponse` parsed from the response body.
 * @param optionalField
 */
function handleResponseResult(optionalField) {
    return function (res) {
        if (lodash_1.default.isNumber(res.status) && res.status >= 200 && res.status < 300) {
            return (
            // If there's an optional field and the body is non-nullish with that property, return it;
            // otherwise return the body if available; if not, return the text; and finally fallback to the entire response.
            (optionalField && res.body && res.body[optionalField] !== undefined ? res.body[optionalField] : res.body) ??
                res.text ??
                res);
        }
        throw errFromResponse(res);
    };
}
/**
 * Extract relevant information from a successful response (that is, a response with an HTTP status code
 * between 200 and 299), but which resulted in an application specific error and use it to construct and
 * throw an `ApiErrorResponse`.
 *
 * @param res
 */
function errFromResponse(res) {
    const message = createResponseErrorString(res);
    const status = res.status;
    const result = res.body;
    const invalidToken = lodash_1.default.has(res.header, 'x-auth-required') && res.header['x-auth-required'] === 'true';
    const needsOtp = res.body?.needsOTP !== undefined;
    return new sdk_core_1.ApiResponseError(message, status, result, invalidToken, needsOtp);
}
/**
 * Handle an error or an error containing an HTTP response and use it to throw a well-formed error object.
 *
 * @param e
 */
function handleResponseError(e) {
    if (e.response) {
        throw errFromResponse(e.response);
    }
    throw e;
}
/**
 * There are many ways a request can fail, and may ways information on that failure can be
 * communicated to the client. This function tries to handle those cases and create a sane error string
 * @param res Response from an HTTP request
 */
function createResponseErrorString(res) {
    let errString = res.status.toString(); // at the very least we'll have the status code
    if (res.body?.error) {
        // this is the case we hope for, where the server gives us a nice error from the JSON body
        errString = res.body.error;
    }
    else if (res.text) {
        // if the response came back as text, we try to parse it as HTML and remove all tags, leaving us
        // just the bare text, which we then trim of excessive newlines and limit to a certain length
        try {
            let sanitizedText = (0, sanitize_html_1.default)(res.text, { allowedTags: [] });
            sanitizedText = sanitizedText.trim();
            sanitizedText = eol_1.default.lf(sanitizedText); // use '\n' for all newlines
            sanitizedText = lodash_1.default.replace(sanitizedText, /\n[ |\t]{1,}\n/g, '\n\n'); // remove the spaces/tabs between newlines
            sanitizedText = lodash_1.default.replace(sanitizedText, /[\n]{3,}/g, '\n\n'); // have at most 2 consecutive newlines
            sanitizedText = sanitizedText.substring(0, 5000); // prevent message from getting too large
            errString = errString + '\n' + sanitizedText; // add it to our existing errString (at this point the more info the better!)
        }
        catch (e) {
            // do nothing, the response's HTML was too wacky to be parsed cleanly
            debug('got error with message "%s" while creating response error string from response: %s', e.message, res.text);
        }
    }
    return errString;
}
/**
 * Serialize request data based on the request content type.
 * If data is already a string, returns it as-is to preserve exact bytes for HMAC.
 * If data is a Buffer, converts to UTF-8 string.
 * If data is an object, serializes it based on Content-Type.
 * @param req
 */
function serializeRequestData(req) {
    const data = req._data;
    if (typeof data === 'string') {
        return data;
    }
    if (Buffer.isBuffer(data)) {
        // Convert Buffer to UTF-8 string and mutate _data to ensure consistency.
        // This is critical: if _data stays as a Buffer, superagent will serialize it
        // as JSON ({"type":"Buffer","data":[...]}), which won't match the HMAC.
        // By mutating to string, superagent sends the exact bytes we hash.
        // This is safe for retries: calling again with a string will hit the early return above.
        const stringData = data.toString('utf8');
        req._data = stringData;
        return stringData;
    }
    if (data === undefined || data === null) {
        return undefined;
    }
    // Serialize object data based on Content-Type
    let contentType = req.get('Content-Type');
    // Parse out just the content type from the header (ignore the charset)
    if (contentType) {
        contentType = contentType.split(';')[0].trim();
    }
    let serialize;
    const serializers = superagent_1.default.serialize;
    if (contentType) {
        if (Object.prototype.hasOwnProperty.call(serializers, contentType)) {
            const candidate = serializers[contentType];
            if (typeof candidate === 'function') {
                serialize = candidate;
            }
        }
        else if (/[\/+]json\b/.test(contentType)) {
            // Fall back to the built-in JSON serializer for JSON-like content types
            if (Object.prototype.hasOwnProperty.call(serializers, 'application/json')) {
                const jsonSerializer = serializers['application/json'];
                if (typeof jsonSerializer === 'function') {
                    serialize = jsonSerializer;
                }
            }
        }
    }
    if (serialize) {
        const serialized = serialize(data);
        req._data = serialized;
        return serialized;
    }
    return undefined;
}
/**
 * Set the superagent query string correctly for browsers or node.
 * @param req
 */
function setRequestQueryString(req) {
    const urlDetails = url_1.default.parse(req.url);
    let queryString;
    const query = req._query;
    const qs = req.qs;
    if (query && query.length > 0) {
        // browser version
        queryString = query.join('&');
        req._query = [];
    }
    else if (qs) {
        // node version
        queryString = querystring_1.default.stringify(qs);
        req.qs = null;
    }
    if (queryString) {
        if (urlDetails.search) {
            urlDetails.search += '&' + queryString;
        }
        else {
            urlDetails.search = '?' + queryString;
        }
        req.url = url_1.default.format(urlDetails);
    }
}
/**
 * Verify that the response received from the server is signed correctly.
 * Right now, it is very permissive with the timestamp variance.
 */
function verifyResponse(bitgo, token, method, req, response, authVersion) {
    // we can't verify the response if we're not authenticated
    if (!req.isV2Authenticated || !req.authenticationToken) {
        return response;
    }
    const verificationResponse = bitgo.verifyResponse({
        url: req.url,
        hmac: response.header.hmac,
        statusCode: response.status,
        text: response.text,
        timestamp: response.header.timestamp,
        token: req.authenticationToken,
        method,
        authVersion,
    });
    if (!verificationResponse.isValid) {
        // calculate the HMAC
        const receivedHmac = response.header.hmac;
        const expectedHmac = verificationResponse.expectedHmac;
        const signatureSubject = verificationResponse.signatureSubject;
        // Log only the first 10 characters of the token to ensure the full token isn't logged.
        const partialBitgoToken = token ? token.substring(0, 10) : '';
        const errorDetails = {
            expectedHmac,
            receivedHmac,
            hmacInput: signatureSubject,
            requestToken: req.authenticationToken,
            bitgoToken: partialBitgoToken,
        };
        debug('Invalid response HMAC: %O', errorDetails);
        throw new sdk_core_1.ApiResponseError('invalid response HMAC, possible man-in-the-middle-attack', 511, errorDetails);
    }
    if (bitgo.getAuthVersion() === 3 && !verificationResponse.isInResponseValidityWindow) {
        const errorDetails = {
            timestamp: response.header.timestamp,
            verificationTime: verificationResponse.verificationTime,
        };
        debug('Server response outside response validity time window: %O', errorDetails);
        throw new sdk_core_1.ApiResponseError('server response outside response validity time window, possible man-in-the-middle-attack', 511, errorDetails);
    }
    return response;
}
