// @bitgo/sdk-api/dist/src/encrypt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.bytesToWord = bytesToWord;
exports.encrypt = encrypt;
exports.decrypt = decrypt;
const sjcl = __importStar(require("@bitgo/sjcl"));
const crypto_1 = require("crypto");
/**
 * convert a 4 element Uint8Array to a 4 byte Number
 *
 * @param bytes
 * @return 4 byte number
 */
function bytesToWord(bytes) {
    if (!(bytes instanceof Uint8Array) || bytes.length !== 4) {
        throw new Error('bytes must be a Uint8Array with length 4');
    }
    return bytes.reduce((num, byte) => num * 0x100 + byte, 0);
}
function encrypt(password, plaintext, options) {
    const salt = options?.salt || (0, crypto_1.randomBytes)(8);
    if (salt.length !== 8) {
        throw new Error(`salt must be 8 bytes`);
    }
    const iv = options?.iv || (0, crypto_1.randomBytes)(16);
    if (iv.length !== 16) {
        throw new Error(`iv must be 16 bytes`);
    }
    const encryptOptions = {
        iter: 10000,
        ks: 256,
        salt: [bytesToWord(salt.slice(0, 4)), bytesToWord(salt.slice(4))],
        iv: [
            bytesToWord(iv.slice(0, 4)),
            bytesToWord(iv.slice(4, 8)),
            bytesToWord(iv.slice(8, 12)),
            bytesToWord(iv.slice(12, 16)),
        ],
    };
    if (options?.adata) {
        encryptOptions.adata = options.adata;
    }
    return sjcl.encrypt(password, plaintext, encryptOptions);
}
function decrypt(password, ciphertext) {
    return sjcl.decrypt(password, ciphertext);
}
