// @bitgo/sdk-api/dist/src/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNetwork = getNetwork;
exports.tryPromise = tryPromise;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const sdk_core_1 = require("@bitgo/sdk-core");
function getNetwork(network) {
    network = network || sdk_core_1.common.getNetwork();
    return utxolib.networks[network];
}
/**
 * A native Promise-based replacement for Bluebird.try
 *
 * This function executes the provided function and returns a Promise.
 * If the function throws synchronously, the exception is caught and returned as a rejected Promise.
 * If the function returns a Promise, that Promise is returned.
 * If the function returns any other value, a resolved Promise with that value is returned.
 *
 * @param {Function} fn - The function to execute
 * @returns {Promise<T>} A promise that resolves with the return value of fn or rejects with any error thrown
 */
function tryPromise(fn) {
    try {
        const result = fn();
        return Promise.resolve(result);
    }
    catch (error) {
        return Promise.reject(error);
    }
}
