// @bitgo/sdk-coin-sui/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKEN_OBJECT_LIMIT = exports.DEFAULT_SCAN_FACTOR = exports.MAX_OBJECT_LIMIT = exports.DEFAULT_GAS_OVERHEAD = exports.DEFAULT_GAS_PRICE = exports.MAX_GAS_BUDGET = exports.MIN_STAKING_THRESHOLD = exports.SIGNATURE_SCHEME_BYTES = exports.SUI_INTENT_BYTES = exports.MAX_GAS_OBJECTS = exports.MAX_COMMAND_ARGS = exports.SER_BUFFER_SIZE = exports.SUI_ADDRESS_LENGTH = exports.DUMMY_SUI_GAS_PRICE = exports.AMOUNT_UNKNOWN_TEXT = exports.UNAVAILABLE_TEXT = void 0;
exports.UNAVAILABLE_TEXT = 'UNAVAILABLE';
exports.AMOUNT_UNKNOWN_TEXT = 'AMOUNT_UNKNOWN';
// Refer to
// https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/messages.rs#L50
exports.DUMMY_SUI_GAS_PRICE = 1000;
exports.SUI_ADDRESS_LENGTH = 32;
exports.SER_BUFFER_SIZE = 8192;
// https://github.com/MystenLabs/sui/blob/e62839aecb192545c3951b6fa0bc84039120b373/crates/sui-protocol-config/src/lib.rs#L1496
exports.MAX_COMMAND_ARGS = 512;
// https://github.com/MystenLabs/sui/blob/e62839aecb192545c3951b6fa0bc84039120b373/crates/sui-protocol-config/src/lib.rs#L1493
exports.MAX_GAS_OBJECTS = 256;
exports.SUI_INTENT_BYTES = Buffer.from([0, 0, 0]);
exports.SIGNATURE_SCHEME_BYTES = [0x00];
exports.MIN_STAKING_THRESHOLD = 1000000000; // 1 SUI
// The maximum gas that is allowed. 0.1 SUI
exports.MAX_GAS_BUDGET = 100000000;
exports.DEFAULT_GAS_PRICE = 1000;
exports.DEFAULT_GAS_OVERHEAD = 1.1;
/*
 * SUI transactions have a limit of 128 KB on the tx size,
 * that corresponds to a tx with max ~1600 objects depending on other transaction details,
 * we chose 1280 as the object limit, keeping in mind other things like multiple recipients etc.
 */
exports.MAX_OBJECT_LIMIT = 1280;
exports.DEFAULT_SCAN_FACTOR = 20; // default number of receive addresses to scan for funds
exports.TOKEN_OBJECT_LIMIT = 512;
