// @bitgo/sdk-coin-sui/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodNames = exports.SuiTransactionType = void 0;
var SuiTransactionType;
(function (SuiTransactionType) {
    SuiTransactionType["Transfer"] = "Transfer";
    SuiTransactionType["AddStake"] = "AddStake";
    SuiTransactionType["WithdrawStake"] = "WithdrawStake";
    SuiTransactionType["CustomTx"] = "CustomTx";
    SuiTransactionType["TokenTransfer"] = "TokenTransfer";
    SuiTransactionType["WalrusStakeWithPool"] = "WalrusStakeWithPool";
    SuiTransactionType["WalrusRequestWithdrawStake"] = "WalrusRequestWithdrawStake";
    SuiTransactionType["WalrusWithdrawStake"] = "WalrusWithdrawStake";
})(SuiTransactionType || (exports.SuiTransactionType = SuiTransactionType = {}));
/**
 * Method names for the transaction method. Names change based on the type of transaction e.g 'request_add_delegation_mul_coin' for the staking transaction
 */
var MethodNames;
(function (MethodNames) {
    /**
     * Add stake to a validator's staking pool.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/sui_system.md#function-request_add_stake
     */
    MethodNames["RequestAddStake"] = "::sui_system::request_add_stake";
    /**
     * Withdraw some portion of a stake from a validator's staking pool.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/sui_system.md#function-request_withdraw_stake
     */
    MethodNames["RequestWithdrawStake"] = "::sui_system::request_withdraw_stake";
    /**
     * Split StakedSui self to two parts, one with principal split_amount, and the remaining principal is left in self.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/staking_pool.md#0x3_staking_pool_split
     */
    MethodNames["StakingPoolSplit"] = "::staking_pool::split";
    /**
     * Transfer ownership of obj to recipient.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/transfer.md#function-public_transfer
     */
    MethodNames["PublicTransfer"] = "::transfer::public_transfer";
    /**
     * Walrus stake with pool.
     *
     * @see https://github.com/MystenLabs/walrus-docs/blob/8ba15d67d7ed0e728077e1600866fddd46fd113b/contracts/walrus/sources/staking.move#L289
     */
    MethodNames["WalrusStakeWithPool"] = "::staking::stake_with_pool";
    /**
     * @see https://github.com/MystenLabs/walrus-docs/blob/9307e66df0ea3f6555cdef78d46aefa62737e216/contracts/walrus/sources/staking.move#L221
     */
    MethodNames["WalrusRequestWithdrawStake"] = "::staking::request_withdraw_stake";
    /**
     * @see https://github.com/MystenLabs/walrus-docs/blob/9307e66df0ea3f6555cdef78d46aefa62737e216/contracts/walrus/sources/staking.move#L231
     */
    MethodNames["WalrusWithdrawStake"] = "::staking::withdraw_stake";
    /**
     * @see https://github.com/MystenLabs/walrus-docs/blob/9307e66df0ea3f6555cdef78d46aefa62737e216/contracts/walrus/sources/staking/staked_wal.move#L143
     */
    MethodNames["WalrusSplitStakedWal"] = "::staked_wal::split";
})(MethodNames || (exports.MethodNames = MethodNames = {}));
