// @bitgo/sdk-coin-sui/dist/src/lib/rpcClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestIgnoreStatusCode = requestIgnoreStatusCode;
exports.makeRPC = makeRPC;
const superagent = __importStar(require("superagent"));
async function request({ url, timeoutMs = undefined, params = {} }) {
    return superagent
        .post(url)
        .send(params)
        .timeout({ deadline: timeoutMs })
        .type('json')
        .set({ Accept: 'application/json' });
}
async function requestIgnoreStatusCode(params) {
    try {
        return await request(params);
    }
    catch (e) {
        if ('response' in e) {
            return e.response;
        }
        throw e;
    }
}
async function makeRPC(url, method, params) {
    const res = await requestIgnoreStatusCode({
        method: 'post',
        url,
        timeoutMs: 3000,
        params: { jsonrpc: '2.0', id: 1, method: method, params: params },
    });
    if (res.body.error) {
        const error = res.body.error;
        throw new Error(`Request to the node failed. Error code: ${error.code}, message: ${error.message}`);
    }
    if (res.statusCode !== 200) {
        throw new Error(`Request to the node failed, Status code: ${res.statusCode}`);
    }
    return res.body.result;
}
