// @bitgo/sdk-coin-sui/dist/src/lib/walrusWithdrawStakeBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalrusWithdrawStakeBuilder = void 0;
const iface_1 = require("./iface");
const statics_1 = require("@bitgo/statics");
const transactionBuilder_1 = require("./transactionBuilder");
const walrusWithdrawStakeTransaction_1 = require("./walrusWithdrawStakeTransaction");
const utils_1 = __importDefault(require("./utils"));
const walrusConfig_1 = require("./resources/walrusConfig");
const sdk_core_1 = require("@bitgo/sdk-core");
const types_1 = require("./mystenlab/types");
const builder_1 = require("./mystenlab/builder");
const assert_1 = __importDefault(require("assert"));
const constants_1 = require("./constants");
class WalrusWithdrawStakeBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new walrusWithdrawStakeTransaction_1.WalrusWithdrawStakeTransaction(_coinConfig);
        // TODO improve mainnet vs. testnet configuration
        this.walrusConfig = _coinConfig.network.type === statics_1.NetworkType.MAINNET ? walrusConfig_1.WALRUS_PROD_CONFIG : walrusConfig_1.WALRUS_TESTNET_CONFIG;
    }
    /**
     * Get staking transaction type
     *
     * @return {TransactionType}
     * @protected
     */
    get transactionType() {
        return utils_1.default.getTransactionType(this.transaction.suiTransaction.type);
    }
    /** @inheritDoc */
    validateTransaction(transaction) {
        if (!transaction.suiTransaction) {
            return;
        }
        this.validateTransactionFields();
    }
    /** @inheritDoc */
    sign(key) {
        this.transaction.setSuiTransaction(this.buildSuiTransaction());
        super.sign(key);
    }
    /**
     * Create a new transaction for requesting coin withdrawal, ready to be signed and executed.
     *
     * @param request - the request object
     */
    requestWithdrawStake(request) {
        this.validateSuiObjectRef(request.stakedWal, 'stakedWal');
        if (request.amount !== undefined) {
            if (!utils_1.default.isValidAmount(request.amount)) {
                throw new Error(`invalid amount: ${request.amount}`);
            }
        }
        this._withdrawStake = request;
        return this;
    }
    /** @inheritDoc */
    fromImplementation(rawTransaction) {
        const tx = new walrusWithdrawStakeTransaction_1.WalrusWithdrawStakeTransaction(this._coinConfig);
        this.validateRawTransaction(rawTransaction);
        tx.fromRawTransaction(rawTransaction);
        this.initBuilder(tx);
        return this.transaction;
    }
    /** @inheritDoc */
    async buildImplementation() {
        this.transaction.setSuiTransaction(this.buildSuiTransaction());
        this.transaction.transactionType(this.transactionType);
        if (this._signer) {
            this.transaction.sign(this._signer);
        }
        this._signatures.forEach((signature) => {
            this.transaction.addSignature(signature.publicKey, signature.signature);
        });
        this.transaction.loadInputsAndOutputs();
        return this.transaction;
    }
    initBuilder(tx) {
        this._transaction = tx;
        if (tx.signature && tx.signature.length > 0) {
            this._signatures = [tx.suiSignature];
        }
        const txData = tx.toJson();
        this.type(tx.suiTransaction.type);
        this.sender(txData.sender);
        this.gasData(txData.gasData);
        const parsed = utils_1.default.getWalrusWithdrawStakeRequests(tx.suiTransaction.tx);
        this.requestWithdrawStake({
            stakedWal: {
                ...parsed.stakedWal,
                objectId: (0, types_1.normalizeSuiObjectId)(parsed.stakedWal.objectId),
                version: Number(parsed.stakedWal.version),
            },
            amount: parsed.amount === undefined ? undefined : Number(parsed.amount),
        });
    }
    validateTransactionFields() {
        (0, assert_1.default)(this._type, new sdk_core_1.BuildTransactionError('type is required before building'));
        (0, assert_1.default)(this._sender, new sdk_core_1.BuildTransactionError('sender is required before building'));
        (0, assert_1.default)(this._withdrawStake.stakedWal, new sdk_core_1.BuildTransactionError('stakedWal object is required before building'));
        (0, assert_1.default)(this._gasData, new sdk_core_1.BuildTransactionError('gasData is required before building'));
        this.validateGasData(this._gasData);
    }
    buildSuiTransaction() {
        this.validateTransactionFields();
        const ptb = new builder_1.TransactionBlock();
        switch (this._type) {
            case iface_1.SuiTransactionType.WalrusRequestWithdrawStake:
                // For request_withdraw, we may need to split the staked WAL into two objects
                if (this._withdrawStake.amount !== undefined) {
                    // This is a partial unstake. Split the StakedWAL
                    const splitStakedWal = ptb.moveCall({
                        target: `${this.walrusConfig.WALRUS_PKG_ID}::staked_wal::split`,
                        arguments: [
                            ptb.object(builder_1.Inputs.ObjectRef(this._withdrawStake.stakedWal)),
                            ptb.pure(Number(this._withdrawStake.amount)),
                        ],
                    });
                    // Request to withdraw the split StakedWAL
                    ptb.moveCall({
                        target: `${this.walrusConfig.WALRUS_PKG_ID}::${this.walrusConfig.WALRUS_STAKING_MODULE_NAME}::request_withdraw_stake`,
                        arguments: [ptb.object(builder_1.Inputs.SharedObjectRef(this.walrusConfig.WALRUS_STAKING_OBJECT)), splitStakedWal],
                    });
                    ptb.transferObjects([splitStakedWal], ptb.object(this._sender));
                }
                else {
                    // This is a full unstake. No need to split, just withdraw the provided StakedWAL.
                    // Also, since there is no newly-created StakedWAL, no need to transferObjects.
                    ptb.moveCall({
                        target: `${this.walrusConfig.WALRUS_PKG_ID}::${this.walrusConfig.WALRUS_STAKING_MODULE_NAME}::request_withdraw_stake`,
                        arguments: [
                            ptb.object(builder_1.Inputs.SharedObjectRef(this.walrusConfig.WALRUS_STAKING_OBJECT)),
                            ptb.object(builder_1.Inputs.ObjectRef(this._withdrawStake.stakedWal)),
                        ],
                    });
                }
                break;
            case iface_1.SuiTransactionType.WalrusWithdrawStake:
                // For the actual withdraw, we are using the already split StakedWAL from above
                // No need to split, just do the withdraw_stake move call
                const unstakedWal = ptb.moveCall({
                    target: `${this.walrusConfig.WALRUS_PKG_ID}::${this.walrusConfig.WALRUS_STAKING_MODULE_NAME}::withdraw_stake`,
                    arguments: [
                        ptb.object(builder_1.Inputs.SharedObjectRef(this.walrusConfig.WALRUS_STAKING_OBJECT)),
                        ptb.object(builder_1.Inputs.ObjectRef(this._withdrawStake.stakedWal)),
                    ],
                });
                ptb.transferObjects([unstakedWal], ptb.object(this._sender));
                break;
            default:
                throw new sdk_core_1.BuildTransactionError(`Invalid transaction type: ${this._type}`);
        }
        const txData = ptb.blockData;
        return {
            type: this._type,
            sender: this._sender,
            tx: {
                inputs: [...txData.inputs],
                transactions: [...txData.transactions],
            },
            gasData: {
                ...this._gasData,
                payment: this._gasData.payment.slice(0, constants_1.MAX_GAS_OBJECTS - 1),
            },
        };
    }
}
exports.WalrusWithdrawStakeBuilder = WalrusWithdrawStakeBuilder;
