// @bitgo/sdk-coin-sui/dist/src/lib/walrusStakingBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalrusStakingBuilder = void 0;
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const iface_1 = require("./iface");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
const assert_1 = __importDefault(require("assert"));
const stakingTransaction_1 = require("./stakingTransaction");
const builder_1 = require("./mystenlab/builder");
const constants_1 = require("./constants");
const walrusConfig_1 = require("./resources/walrusConfig");
class WalrusStakingBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new stakingTransaction_1.StakingTransaction(_coinConfig);
        // TODO improve mainnet vs. testnet configuration
        this.walrusConfig = _coinConfig.network.type === statics_1.NetworkType.MAINNET ? walrusConfig_1.WALRUS_PROD_CONFIG : walrusConfig_1.WALRUS_TESTNET_CONFIG;
    }
    /**
     * Build a MoveCall transaction ready to be signed and executed.
     *
     * @returns {BitGoSuiTransaction} an unsigned Sui transaction
     */
    buildStakeTransaction() {
        return {
            type: iface_1.SuiTransactionType.WalrusStakeWithPool,
            sender: this._sender,
            tx: {
                inputs: [],
                transactions: [],
            },
            gasData: this._gasData,
        };
    }
    /**
     * Get staking transaction type
     *
     * @return {TransactionType}
     * @protected
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingAdd;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction.suiTransaction) {
            return;
        }
        this.validateTransactionFields();
    }
    /** @inheritdoc */
    sign(key) {
        this.transaction.setSuiTransaction(this.buildSuiTransaction());
        super.sign(key);
    }
    /**
     * Create a new transaction for staking coins ready to be signed and executed.
     *
     * @param {RequestWalrusStakeWithPool[]} request: a list of staking request
     */
    stake(request) {
        request.forEach((req) => {
            utils_1.default.validateAddress(req.validatorAddress, 'validatorAddress');
            (0, assert_1.default)(utils_1.default.isValidAmount(req.amount), 'Invalid recipient amount');
            if (this._sender === req.validatorAddress) {
                throw new sdk_core_1.BuildTransactionError('Sender address cannot be the same as the Staking address');
            }
        });
        this._stakeWithPoolTx = request;
        return this;
    }
    /**
     * Set the $WAL objects to be used for staking.
     *
     * @param tokens The WAL objects to be used
     * @returns this
     */
    inputObjects(inputObjects) {
        this.validateInputObjects(inputObjects);
        this._inputObjects = inputObjects;
        return this;
    }
    validateInputObjects(inputObjects) {
        (0, assert_1.default)(inputObjects && inputObjects.length > 0, new sdk_core_1.BuildTransactionError('input objects required before building'));
        inputObjects.forEach((inputObject) => {
            this.validateSuiObjectRef(inputObject, 'input object');
        });
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = new stakingTransaction_1.StakingTransaction(this._coinConfig);
        this.validateRawTransaction(rawTransaction);
        tx.fromRawTransaction(rawTransaction);
        this.initBuilder(tx);
        return this.transaction;
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.setSuiTransaction(this.buildSuiTransaction());
        this.transaction.transactionType(this.transactionType);
        if (this._signer) {
            this.transaction.sign(this._signer);
        }
        this._signatures.forEach((signature) => {
            this.transaction.addSignature(signature.publicKey, signature.signature);
        });
        this.transaction.loadInputsAndOutputs();
        return this.transaction;
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {StakingTransaction} tx the transaction data
     */
    initBuilder(tx) {
        this._transaction = tx;
        if (tx.signature && tx.signature.length > 0) {
            this._signatures = [tx.suiSignature];
        }
        const txData = tx.toJson();
        this.type(iface_1.SuiTransactionType.WalrusStakeWithPool);
        this.sender(txData.sender);
        this.gasData(txData.gasData);
        const requests = utils_1.default.getWalrusStakeWithPoolRequests(tx.suiTransaction.tx);
        this.stake(requests);
        (0, assert_1.default)(txData.inputObjects);
        this.inputObjects(txData.inputObjects);
    }
    /**
     * Validates all fields are defined
     */
    validateTransactionFields() {
        (0, assert_1.default)(this._type, new sdk_core_1.BuildTransactionError('type is required before building'));
        (0, assert_1.default)(this._sender, new sdk_core_1.BuildTransactionError('sender is required before building'));
        this._stakeWithPoolTx.forEach((req) => {
            (0, assert_1.default)(req.validatorAddress, new sdk_core_1.BuildTransactionError('validator address is required before building'));
            (0, assert_1.default)(req.amount, new sdk_core_1.BuildTransactionError('staking amount is required before building'));
        });
        (0, assert_1.default)(this._gasData, new sdk_core_1.BuildTransactionError('gasData is required before building'));
        this.validateGasData(this._gasData);
        this.validateInputObjects(this._inputObjects);
    }
    /**
     * Build SuiTransaction
     *
     * @return {BitGoSuiTransaction<MoveCallTx>}
     * @protected
     */
    buildSuiTransaction() {
        this.validateTransactionFields();
        const programmableTxBuilder = new builder_1.TransactionBlock();
        switch (this._type) {
            case iface_1.SuiTransactionType.WalrusStakeWithPool:
                const inputObjects = this._inputObjects.map((token) => programmableTxBuilder.object(builder_1.Inputs.ObjectRef(token)));
                const mergedObject = inputObjects.shift();
                if (inputObjects.length > 0) {
                    programmableTxBuilder.mergeCoins(mergedObject, inputObjects);
                }
                // Create a new coin with staking balance, based on the coins used as gas payment.
                const stakedWals = this._stakeWithPoolTx.map((req) => {
                    const splitObject = programmableTxBuilder.splitCoins(mergedObject, [
                        programmableTxBuilder.pure(Number(req.amount)),
                    ]);
                    // Stake the split coin to a specific validator address.
                    return programmableTxBuilder.moveCall({
                        target: `${this.walrusConfig.WALRUS_PKG_ID}::${this.walrusConfig.WALRUS_STAKING_MODULE_NAME}::${this.walrusConfig.WALRUS_STAKE_WITH_POOL_FUN_NAME}`,
                        arguments: [
                            programmableTxBuilder.object(builder_1.Inputs.SharedObjectRef(this.walrusConfig.WALRUS_STAKING_OBJECT)),
                            splitObject,
                            programmableTxBuilder.object(req.validatorAddress),
                        ],
                    });
                });
                programmableTxBuilder.transferObjects(stakedWals, programmableTxBuilder.object(this._sender));
                break;
            default:
                throw new sdk_core_1.InvalidTransactionError(`unsupported target method`);
        }
        const txData = programmableTxBuilder.blockData;
        return {
            type: this._type,
            sender: this._sender,
            tx: {
                inputs: [...txData.inputs],
                transactions: [...txData.transactions],
            },
            gasData: {
                ...this._gasData,
                payment: this._gasData.payment.slice(0, constants_1.MAX_GAS_OBJECTS - 1),
            },
        };
    }
}
exports.WalrusStakingBuilder = WalrusStakingBuilder;
