// @bitgo/sdk-coin-rune/dist/test/unit/rune.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const mocha_1 = require("mocha");
const sinon_1 = __importDefault(require("sinon"));
const src_1 = require("../../src");
const utils_1 = require("../../src/lib/utils");
const rune_1 = require("../resources/rune");
const trune_1 = require("../resources/trune");
const bech32 = require('bech32-buffer');
const should = require("should");
describe('Rune', function () {
    let bitgo;
    let rune;
    let trune;
    let mainnetUtils;
    let testnetUtils;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('thorchain:rune', src_1.Rune.createInstance);
        bitgo.safeRegister('tthorchain:rune', src_1.Trune.createInstance);
        bitgo.initializeTestVars();
        rune = bitgo.coin('thorchain:rune');
        trune = bitgo.coin('tthorchain:rune');
        mainnetUtils = new utils_1.RuneUtils(statics_1.NetworkType.MAINNET);
        testnetUtils = new utils_1.RuneUtils(statics_1.NetworkType.TESTNET);
    });
    it('should return the right info', function () {
        rune.getChain().should.equal('thorchain:rune');
        rune.getFamily().should.equal('thor');
        rune.getFullName().should.equal('Rune');
        rune.getBaseFactor().should.equal(1e8);
        trune.getChain().should.equal('tthorchain:rune');
        trune.getFamily().should.equal('thor');
        trune.getFullName().should.equal('Testnet Rune');
        trune.getBaseFactor().should.equal(1e8);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const mainnetAddressDetails = rune.getAddressDetails(rune_1.mainnetAddress.noMemoIdAddress);
            mainnetAddressDetails.address.should.equal(rune_1.mainnetAddress.noMemoIdAddress);
            should.not.exist(mainnetAddressDetails.memoId);
            const testnetAddressDetails = trune.getAddressDetails(trune_1.testnetAddress.noMemoIdAddress);
            testnetAddressDetails.address.should.equal(trune_1.testnetAddress.noMemoIdAddress);
            should.not.exist(testnetAddressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const mainnetAddressDetails = rune.getAddressDetails(rune_1.mainnetAddress.validMemoIdAddress);
            mainnetAddressDetails.address.should.equal(rune_1.mainnetAddress.validMemoIdAddress.split('?')[0]);
            mainnetAddressDetails.memoId.should.equal('2');
            const testnetAddressDetails = rune.getAddressDetails(trune_1.testnetAddress.validMemoIdAddress);
            testnetAddressDetails.address.should.equal(trune_1.testnetAddress.validMemoIdAddress.split('?')[0]);
            testnetAddressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                rune.getAddressDetails(rune_1.mainnetAddress.multipleMemoIdAddress);
            }).should.throw();
            (() => {
                trune.getAddressDetails(trune_1.testnetAddress.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const mainnetReceiveAddress = {
                address: 'thor1fwk9jl6kfflurj9p0wt098kxl02gle4yhnm687?memoId=7',
                coinSpecific: {
                    rootAddress: 'thor1fwk9jl6kfflurj9p0wt098kxl02gle4yhnm687',
                    memoID: '7',
                },
            };
            const testnetReceiveAddress = {
                address: 'sthor19phfqh3ce3nnjhh0cssn433nydq9shx76s8qgg?memoId=7',
                coinSpecific: {
                    rootAddress: 'sthor19phfqh3ce3nnjhh0cssn433nydq9shx76s8qgg',
                    memoID: '7',
                },
            };
            const isValidMainnetReceiveAddress = await rune.isWalletAddress(mainnetReceiveAddress);
            const isValidTestnetReceiveAddress = await trune.isWalletAddress(testnetReceiveAddress);
            isValidMainnetReceiveAddress.should.equal(true);
            isValidTestnetReceiveAddress.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(mainnetUtils.isValidAddress(rune_1.mainnetAddress.address1), true);
            should.equal(mainnetUtils.isValidAddress(rune_1.mainnetAddress.address2), true);
            should.equal(mainnetUtils.isValidAddress(rune_1.mainnetAddress.address3), false);
            should.equal(mainnetUtils.isValidAddress(rune_1.mainnetAddress.address4), false);
            should.equal(mainnetUtils.isValidAddress('dfjk35y'), false);
            should.equal(mainnetUtils.isValidAddress(undefined), false);
            should.equal(mainnetUtils.isValidAddress(''), false);
            should.equal(mainnetUtils.isValidAddress(rune_1.mainnetAddress.validMemoIdAddress), true);
            should.equal(mainnetUtils.isValidAddress(rune_1.mainnetAddress.multipleMemoIdAddress), false);
            should.equal(testnetUtils.isValidAddress(trune_1.testnetAddress.address1), true);
            should.equal(testnetUtils.isValidAddress(trune_1.testnetAddress.address2), true);
            should.equal(testnetUtils.isValidAddress(trune_1.testnetAddress.address3), false);
            should.equal(testnetUtils.isValidAddress(trune_1.testnetAddress.address4), false);
            should.equal(testnetUtils.isValidAddress('dfjk35y'), false);
            should.equal(testnetUtils.isValidAddress(undefined), false);
            should.equal(testnetUtils.isValidAddress(''), false);
            should.equal(testnetUtils.isValidAddress(trune_1.testnetAddress.validMemoIdAddress), true);
            should.equal(testnetUtils.isValidAddress(trune_1.testnetAddress.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(mainnetUtils.isValidValidatorAddress(rune_1.mainnetAddress.validatorAddress1), true);
            should.equal(mainnetUtils.isValidValidatorAddress(rune_1.mainnetAddress.validatorAddress2), true);
            should.equal(mainnetUtils.isValidValidatorAddress(rune_1.mainnetAddress.validatorAddress3), false);
            should.equal(mainnetUtils.isValidValidatorAddress(rune_1.mainnetAddress.validatorAddress4), false);
            should.equal(mainnetUtils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(mainnetUtils.isValidValidatorAddress(undefined), false);
            should.equal(mainnetUtils.isValidValidatorAddress(''), false);
            should.equal(testnetUtils.isValidValidatorAddress(trune_1.testnetAddress.validatorAddress1), true);
            should.equal(testnetUtils.isValidValidatorAddress(trune_1.testnetAddress.validatorAddress2), true);
            should.equal(testnetUtils.isValidValidatorAddress(trune_1.testnetAddress.validatorAddress3), false);
            should.equal(testnetUtils.isValidValidatorAddress(trune_1.testnetAddress.validatorAddress4), false);
            should.equal(testnetUtils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(testnetUtils.isValidValidatorAddress(undefined), false);
            should.equal(testnetUtils.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: trune_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: trune_1.TEST_SEND_TX.recipient,
                        amount: trune_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await trune.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify transaction having recipient address has memo', async function () {
            const txPrebuild = {
                txHex: trune_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: trune_1.TEST_SEND_TX.recipient + '?memoId=2',
                        amount: trune_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await trune.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await trune
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await trune.explainTransaction({
                txHex: trune_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: trune_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: bech32.decode(trune_1.TEST_SEND_TX.recipient).data,
                        amount: trune_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: trune_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: trune_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await trune.explainTransaction({
                txHex: trune_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: trune_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: bech32.decode(trune_1.TEST_TX_WITH_MEMO.recipient).data,
                        amount: trune_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: trune_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: trune_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: trune_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await trune.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await trune.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        it('should parse a transfer transaction', async function () {
            const transferInputsResponse = {
                address: bech32.decode(trune_1.TEST_SEND_TX.recipient).data,
                amount: new bignumber_js_1.default(trune_1.TEST_SEND_TX.sendAmount).plus(trune_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
            };
            const transferOutputsResponse = {
                address: bech32.decode(trune_1.TEST_SEND_TX.recipient).data,
                amount: trune_1.TEST_SEND_TX.sendAmount,
            };
            const parsedTransaction = await trune.parseTransaction({ txHex: trune_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Rune.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await trune
                .parseTransaction({ txHex: trune_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
    describe('Recover transaction: success path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const coin = statics_1.coins.get('tthorchain:rune');
        const testBalance = [
            {
                denom: 'rune',
                amount: '150000000',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'thorchain-stagenet-2';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.Trune.prototype, 'getAccountBalance');
            accountBalance.withArgs(trune_1.wrwUser.senderAddress).resolves(testBalance);
            const accountDetails = sandBox.stub(src_1.Trune.prototype, 'getAccountDetails');
            accountDetails.withArgs(trune_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Trune.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should recover funds for non-bitgo recoveries', async function () {
            const res = await trune.recover({
                userKey: trune_1.wrwUser.userPrivateKey,
                backupKey: trune_1.wrwUser.backupPrivateKey,
                bitgoKey: trune_1.wrwUser.bitgoPublicKey,
                walletPassphrase: trune_1.wrwUser.walletPassphrase,
                recoveryDestination: trune_1.wrwUser.destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(trune.getAccountBalance);
            sandBox.assert.calledOnce(trune.getAccountDetails);
            sandBox.assert.calledOnce(trune.getChainId);
            const truneTxn = new abstract_cosmos_1.CosmosTransaction(coin, testnetUtils);
            truneTxn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const truneTxnJson = truneTxn.toJson();
            const sendMessage = truneTxnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const actualBalance = balance.minus('2000000'); // native rune transaction fees
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
        });
    });
    describe('Recover transaction: failure path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const testZeroBalance = [
            {
                denom: 'rune',
                amount: '0',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'thorchain-stagenet-2';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.Trune.prototype, 'getAccountBalance');
            accountBalance.withArgs(trune_1.wrwUser.senderAddress).resolves(testZeroBalance);
            const accountDetails = sandBox.stub(src_1.Trune.prototype, 'getAccountDetails');
            accountDetails.withArgs(trune_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Trune.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should throw error if backupkey is not present', async function () {
            await trune
                .recover({
                userKey: trune_1.wrwUser.userPrivateKey,
                bitgoKey: trune_1.wrwUser.bitgoPublicKey,
                walletPassphrase: trune_1.wrwUser.walletPassphrase,
                recoveryDestination: trune_1.wrwUser.destinationAddress,
            })
                .should.rejectedWith('missing backupKey');
        });
        it('should throw error if userkey is not present', async function () {
            await trune
                .recover({
                backupKey: trune_1.wrwUser.backupPrivateKey,
                bitgoKey: trune_1.wrwUser.bitgoPublicKey,
                walletPassphrase: trune_1.wrwUser.walletPassphrase,
                recoveryDestination: trune_1.wrwUser.destinationAddress,
            })
                .should.rejectedWith('missing userKey');
        });
        it('should throw error if wallet passphrase is not present', async function () {
            await trune
                .recover({
                userKey: trune_1.wrwUser.userPrivateKey,
                backupKey: trune_1.wrwUser.backupPrivateKey,
                bitgoKey: trune_1.wrwUser.bitgoPublicKey,
                recoveryDestination: trune_1.wrwUser.destinationAddress,
            })
                .should.rejectedWith('missing wallet passphrase');
        });
        it('should throw error if there is no balance', async function () {
            await trune
                .recover({
                userKey: trune_1.wrwUser.userPrivateKey,
                backupKey: trune_1.wrwUser.backupPrivateKey,
                bitgoKey: trune_1.wrwUser.bitgoPublicKey,
                walletPassphrase: trune_1.wrwUser.walletPassphrase,
                recoveryDestination: trune_1.wrwUser.destinationAddress,
            })
                .should.rejectedWith('Did not have enough funds to recover');
        });
    });
});
