// @bitgo/public-types/dist/src/schema/mpcv2/common/mpcv2BroadcastMessage.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MPCv2BroadcastMessageWithSignatureR = exports.MPCv2BroadcastMessage = void 0;
const t = __importStar(require("io-ts"));
const mpcv2PartyFromStringOrNumber_1 = require("./mpcv2PartyFromStringOrNumber");
exports.MPCv2BroadcastMessage = t.type({
    from: mpcv2PartyFromStringOrNumber_1.MPCv2PartyFromStringOrNumber,
    message: t.string,
    signature: t.string,
}, "MPCv2BroadcastMessage");
exports.MPCv2BroadcastMessageWithSignatureR = t.intersection([
    exports.MPCv2BroadcastMessage,
    t.type({
        signatureR: t.string,
    }),
]);
