// @bitgo/wasm-utxo/dist/cjs/js/fixedScriptWallet/Dimensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dimensions = void 0;
const wasm_utxo_js_1 = require("../wasm/wasm_utxo.js");
const address_js_1 = require("../address.js");
/**
 * Dimensions class for estimating transaction virtual size.
 *
 * Tracks weight internally with min/max bounds to handle ECDSA signature variance.
 * Schnorr signatures have no variance (always 64 bytes).
 *
 * This is a thin wrapper over the WASM implementation.
 */
class Dimensions {
    _wasm;
    constructor(_wasm) {
        this._wasm = _wasm;
    }
    /**
     * Create empty dimensions (zero weight)
     */
    static empty() {
        return new Dimensions(wasm_utxo_js_1.WasmDimensions.empty());
    }
    /**
     * Create dimensions from a BitGoPsbt
     *
     * Parses PSBT inputs and outputs to compute weight bounds without
     * requiring wallet keys. Input types are detected from BIP32 derivation
     * paths stored in the PSBT.
     */
    static fromPsbt(psbt) {
        return new Dimensions(wasm_utxo_js_1.WasmDimensions.from_psbt(psbt.wasm));
    }
    /**
     * Create dimensions for a single input
     *
     * @param params - Either `{ chain, signPath? }` or `{ scriptType }`
     * @param options - Optional settings like `{ utxolibCompat: true }` for @bitgo/unspents-compatible sizing
     */
    static fromInput(params, options) {
        const compat = options?.utxolibCompat;
        if ("scriptType" in params) {
            return new Dimensions(wasm_utxo_js_1.WasmDimensions.from_input_script_type(params.scriptType, compat));
        }
        return new Dimensions(wasm_utxo_js_1.WasmDimensions.from_input(params.chain, params.signPath?.signer, params.signPath?.cosigner, compat));
    }
    static fromOutput(params, network) {
        if (typeof params === "string") {
            if (network === undefined) {
                throw new Error("network is required when passing an address string");
            }
            const script = (0, address_js_1.toOutputScriptWithCoin)(params, network);
            return new Dimensions(wasm_utxo_js_1.WasmDimensions.from_output_script_length(script.length));
        }
        if (typeof params === "object" && "scriptType" in params) {
            return new Dimensions(wasm_utxo_js_1.WasmDimensions.from_output_script_type(params.scriptType));
        }
        // Both Uint8Array and { length: number } have .length
        return new Dimensions(wasm_utxo_js_1.WasmDimensions.from_output_script_length(params.length));
    }
    /**
     * Combine with another Dimensions instance
     */
    plus(other) {
        return new Dimensions(this._wasm.plus(other._wasm));
    }
    /**
     * Multiply dimensions by a scalar
     */
    times(n) {
        return new Dimensions(this._wasm.times(n));
    }
    /**
     * Whether any inputs are segwit (affects overhead calculation)
     */
    get hasSegwit() {
        return this._wasm.has_segwit();
    }
    /**
     * Get total weight (min or max)
     * @param size - "min" or "max", defaults to "max"
     */
    getWeight(size = "max") {
        return this._wasm.get_weight(size);
    }
    /**
     * Get virtual size (min or max)
     * @param size - "min" or "max", defaults to "max"
     */
    getVSize(size = "max") {
        return this._wasm.get_vsize(size);
    }
    /**
     * Get input weight only (min or max)
     * @param size - "min" or "max", defaults to "max"
     */
    getInputWeight(size = "max") {
        return this._wasm.get_input_weight(size);
    }
    /**
     * Get input virtual size (min or max)
     * @param size - "min" or "max", defaults to "max"
     */
    getInputVSize(size = "max") {
        return this._wasm.get_input_vsize(size);
    }
    /**
     * Get output weight
     */
    getOutputWeight() {
        return this._wasm.get_output_weight();
    }
    /**
     * Get output virtual size
     */
    getOutputVSize() {
        return this._wasm.get_output_vsize();
    }
}
exports.Dimensions = Dimensions;
