// @bitgo/sdk-coin-cronos/dist/test/unit/cronos.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should = require("should");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sinon_1 = __importDefault(require("sinon"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const index_1 = require("../../src/index");
const utils_1 = require("../../src/lib/utils");
const cronos_1 = require("../resources/cronos");
const statics_1 = require("@bitgo/statics");
const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
bitgo.safeRegister('tcronos', index_1.Tcronos.createInstance);
describe('Cronos', function () {
    let bitgo;
    let basecoin;
    let utils;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('cronos', index_1.Cronos.createInstance);
        bitgo.safeRegister('tcronos', index_1.Tcronos.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('tcronos');
        utils = new utils_1.Utils(statics_1.NetworkType.TESTNET);
    });
    it('should return the right info', function () {
        const cronos = bitgo.coin('cronos');
        const tcronos = bitgo.coin('tcronos');
        cronos.getChain().should.equal('cronos');
        cronos.getFamily().should.equal('cronos');
        cronos.getFullName().should.equal('Cronos POS');
        cronos.getBaseFactor().should.equal(1e8);
        tcronos.getChain().should.equal('tcronos');
        tcronos.getFamily().should.equal('cronos');
        tcronos.getFullName().should.equal('Testnet Cronos POS');
        tcronos.getBaseFactor().should.equal(1e8);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const addressDetails = basecoin.getAddressDetails(cronos_1.address.noMemoIdAddress);
            addressDetails.address.should.equal(cronos_1.address.noMemoIdAddress);
            should.not.exist(addressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const addressDetails = basecoin.getAddressDetails(cronos_1.address.validMemoIdAddress);
            addressDetails.address.should.equal(cronos_1.address.validMemoIdAddress.split('?')[0]);
            addressDetails.memoId.should.equal('2');
        });
        it('should throw on invalid memo id address', () => {
            (() => {
                basecoin.getAddressDetails(cronos_1.address.invalidMemoIdAddress);
            }).should.throw();
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                basecoin.getAddressDetails(cronos_1.address.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const receiveAddress = {
                address: 'tcro1e9rxy3j3wph0lqjxr0ynu0t3zjfnhc0csyldtl?memoId=7',
                coinSpecific: {
                    rootAddress: 'tcro1e9rxy3j3wph0lqjxr0ynu0t3zjfnhc0csyldtl',
                    memoID: '7',
                },
            };
            const isValid = await basecoin.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(utils.isValidAddress(cronos_1.address.address1), true);
            should.equal(utils.isValidAddress(cronos_1.address.address2), true);
            should.equal(utils.isValidAddress(cronos_1.address.address3), false);
            should.equal(utils.isValidAddress(cronos_1.address.address4), false);
            should.equal(utils.isValidAddress('dfjk35y'), false);
            should.equal(utils.isValidAddress(undefined), false);
            should.equal(utils.isValidAddress(''), false);
            should.equal(utils.isValidAddress(cronos_1.address.validMemoIdAddress), true);
            should.equal(utils.isValidAddress(cronos_1.address.invalidMemoIdAddress), false);
            should.equal(utils.isValidAddress(cronos_1.address.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(utils.isValidValidatorAddress(cronos_1.address.validatorAddress1), true);
            should.equal(utils.isValidValidatorAddress(cronos_1.address.validatorAddress2), true);
            should.equal(utils.isValidValidatorAddress(cronos_1.address.validatorAddress3), false);
            should.equal(utils.isValidValidatorAddress(cronos_1.address.validatorAddress4), false);
            should.equal(utils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(utils.isValidValidatorAddress(undefined), false);
            should.equal(utils.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: cronos_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: cronos_1.TEST_SEND_TX.recipient,
                        amount: cronos_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify sendMany transaction', async function () {
            const txPrebuild = {
                txHex: cronos_1.TEST_SEND_MANY_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: 'tcro1rhs402vnjf7369yyjkk0nrskupmfl4yxpnaahj',
                        amount: '10000',
                    },
                    {
                        address: 'tcro1u73cmemdgd5qx06cmtehr3z5pgswtar65f9sp5',
                        amount: '10000',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: cronos_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: cronos_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: cronos_1.TEST_SEND_TX.recipient,
                        amount: cronos_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: cronos_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: cronos_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain sendMany transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: cronos_1.TEST_SEND_MANY_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: cronos_1.TEST_SEND_MANY_TX.hash,
                outputs: [
                    {
                        address: 'tcro1rhs402vnjf7369yyjkk0nrskupmfl4yxpnaahj',
                        amount: '10000',
                    },
                    {
                        address: 'tcro1u73cmemdgd5qx06cmtehr3z5pgswtar65f9sp5',
                        amount: '10000',
                    },
                ],
                outputAmount: '20000',
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: cronos_1.TEST_SEND_MANY_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: cronos_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: cronos_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: cronos_1.TEST_TX_WITH_MEMO.to,
                        amount: cronos_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: cronos_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: cronos_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: cronos_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await basecoin.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await basecoin.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        const transferInputsResponse = {
            address: cronos_1.TEST_SEND_TX.recipient,
            amount: new bignumber_js_1.default(cronos_1.TEST_SEND_TX.sendAmount).plus(cronos_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
        };
        const transferOutputsResponse = {
            address: cronos_1.TEST_SEND_TX.recipient,
            amount: cronos_1.TEST_SEND_TX.sendAmount,
        };
        it('should parse a transfer transaction', async function () {
            const parsedTransaction = await basecoin.parseTransaction({ txHex: cronos_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(index_1.Cronos.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await basecoin
                .parseTransaction({ txHex: cronos_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
});
