// @bitgo/sdk-coin-celo/dist/test/unit/transactionBuilder/send.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const ethUtil = __importStar(require("ethereumjs-util"));
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const sdk_core_1 = require("@bitgo/sdk-core");
const getBuilder_1 = require("../getBuilder");
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/celo"));
const sdk_coin_eth_1 = require("@bitgo/sdk-coin-eth");
describe('Send transaction', function () {
    let txBuilder;
    const initTxBuilder = () => {
        txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
        txBuilder.fee({
            fee: '1000000000',
            gasLimit: '12100000',
        });
        txBuilder.counter(2);
        txBuilder.type(sdk_core_1.TransactionType.Send);
    };
    const key = testData.KEYPAIR_PRV.getKeys().prv;
    const getOperationHash = function (tx) {
        const { data } = tx.toJson();
        const { tokenContractAddress, expireTime, sequenceId, amount, to } = (0, sdk_coin_eth_1.decodeTransferData)(data);
        const operationParams = [
            ['string', 'address', 'uint', 'address', 'uint', 'uint'],
            [
                'CELO-ERC20',
                new ethUtil.BN(ethUtil.stripHexPrefix(to), 16),
                amount,
                new ethUtil.BN(ethUtil.stripHexPrefix(tokenContractAddress || ''), 16),
                expireTime,
                sequenceId,
            ],
        ];
        return ethereumjs_abi_1.default.soliditySHA3(...operationParams).toString('hex');
    };
    describe('should sign and build', () => {
        it('a send celo transaction', async () => {
            const recipient = '0x19645032c7f1533395d44a629462e751084d3e4c';
            const contractAddress = '0x8f977e912ef500548a0c3be6ddde9899f1199b81';
            const amount = '1000000000';
            initTxBuilder();
            txBuilder.contract(contractAddress);
            txBuilder
                .transfer()
                .coin('tcelo')
                .amount(amount)
                .to(recipient)
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            txBuilder.sign({ key: testData.PRIVATE_KEY });
            const tx = await txBuilder.build();
            should_1.default.equal(tx.signature.length, 2);
            should_1.default.equal(tx.inputs.length, 1);
            should_1.default.equal(tx.inputs[0].address, contractAddress);
            should_1.default.equal(tx.inputs[0].value, amount);
            should_1.default.equal(tx.inputs[0].coin, 'tcelo');
            should_1.default.equal(tx.outputs.length, 1);
            should_1.default.equal(tx.outputs[0].address, recipient);
            should_1.default.equal(tx.outputs[0].value, amount);
            should_1.default.equal(tx.outputs[0].coin, 'tcelo');
            const { signature } = (0, sdk_coin_eth_1.decodeTransferData)(tx.toJson().data);
            const operationHash = getOperationHash(tx);
            const { v, r, s } = ethUtil.fromRpcSig(signature);
            const senderPubKey = ethUtil.ecrecover(Buffer.from(ethUtil.padToEven(operationHash || ''), 'hex'), v, r, s);
            const senderAddress = ethUtil.pubToAddress(senderPubKey);
            const senderKey = new src_1.KeyPair({ prv: testData.PRIVATE_KEY });
            ethUtil.bufferToHex(senderAddress).should.equal(senderKey.getAddress());
        });
        it('a send token transaction', async () => {
            const recipient = '0x19645032c7f1533395d44a629462e751084d3e4c';
            const contractAddress = '0x8f977e912ef500548a0c3be6ddde9899f1199b81';
            const amount = '1000000000';
            initTxBuilder();
            txBuilder.contract(contractAddress);
            txBuilder
                .transfer()
                .coin('tcusd')
                .amount(amount)
                .to(recipient)
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            txBuilder.sign({ key: testData.PRIVATE_KEY });
            const tx = await txBuilder.build();
            should_1.default.equal(tx.toBroadcastFormat(), testData.SEND_TOKEN_TX_BROADCAST);
            should_1.default.equal(tx.signature.length, 2);
            should_1.default.equal(tx.inputs.length, 1);
            should_1.default.equal(tx.inputs[0].address, contractAddress);
            should_1.default.equal(tx.inputs[0].value, amount);
            should_1.default.equal(tx.inputs[0].coin, 'tcusd');
            should_1.default.equal(tx.outputs.length, 1);
            should_1.default.equal(tx.outputs[0].address, recipient);
            should_1.default.equal(tx.outputs[0].value, amount);
            should_1.default.equal(tx.outputs[0].coin, 'tcusd');
            const { signature } = (0, sdk_coin_eth_1.decodeTransferData)(tx.toJson().data);
            const operationHash = getOperationHash(tx);
            const { v, r, s } = ethUtil.fromRpcSig(signature);
            const senderPubKey = ethUtil.ecrecover(Buffer.from(ethUtil.padToEven(operationHash || ''), 'hex'), v, r, s);
            const senderAddress = ethUtil.pubToAddress(senderPubKey);
            const senderKey = new src_1.KeyPair({ prv: testData.PRIVATE_KEY });
            ethUtil.bufferToHex(senderAddress).should.equal(senderKey.getAddress());
        });
        it('a send token transactions from serialized', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.from(testData.SEND_TOKEN_TX_BROADCAST);
            const tx = await txBuilder.build();
            should_1.default.equal(tx.toBroadcastFormat(), testData.SEND_TOKEN_TX_BROADCAST);
            const { signature } = (0, sdk_coin_eth_1.decodeTransferData)(tx.toJson().data);
            const operationHash = getOperationHash(tx);
            const { v, r, s } = ethUtil.fromRpcSig(signature);
            const senderPubKey = ethUtil.ecrecover(Buffer.from(ethUtil.padToEven(operationHash || ''), 'hex'), v, r, s);
            const senderAddress = ethUtil.pubToAddress(senderPubKey);
            const senderKey = new src_1.KeyPair({ prv: testData.PRIVATE_KEY });
            ethUtil.bufferToHex(senderAddress).should.equal(senderKey.getAddress());
        });
        it('a half signed transaction', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.from(testData.HALF_SIGNED_TX_SEND);
            txBuilder.transfer().key(key);
            const tx = await txBuilder.build();
            const { signature } = (0, sdk_coin_eth_1.decodeTransferData)(tx.toJson().data);
            const operationHash = getOperationHash(tx);
            const { v, r, s } = ethUtil.fromRpcSig(signature);
            const senderPubKey = ethUtil.ecrecover(Buffer.from(ethUtil.padToEven(operationHash || ''), 'hex'), v, r, s);
            const senderAddress = ethUtil.pubToAddress(senderPubKey);
            const senderKey = new src_1.KeyPair({ prv: testData.PRIVATE_KEY });
            ethUtil.bufferToHex(senderAddress).should.equal(senderKey.getAddress());
        });
    });
    describe('Should build without sign', () => {
        it('a send token transaction without from', async () => {
            const recipient = '0x19645032c7f1533395d44a629462e751084d3e4c';
            const contractAddress = '0x8f977e912ef500548a0c3be6ddde9899f1199b81';
            const amount = '1000000000';
            initTxBuilder();
            txBuilder.contract(contractAddress);
            txBuilder
                .transfer()
                .coin('tcusd')
                .amount(amount)
                .to(recipient)
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.from, undefined);
        });
        it('should build txn with eip1559 for tcelo', async function () {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.fee({
                fee: '1000000000',
                gasLimit: '12100000',
                eip1559: {
                    maxFeePerGas: '7593123',
                    maxPriorityFeePerGas: '150',
                },
            });
            txBuilder.counter(2);
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.contract('0x8f977e912ef500548a0c3be6ddde9899f1199b81');
            txBuilder
                .transfer()
                .coin('tcusd')
                .amount('1000000000')
                .to('0x19645032c7f1533395d44a629462e751084d3e4c')
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            txBuilder.sign({ key: testData.PRIVATE_KEY });
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.gasLimit, '12100000');
            should_1.default.equal(txJson._type, 'EIP1559');
            should_1.default.equal(txJson.maxFeePerGas, '7593123');
            should_1.default.equal(txJson.maxPriorityFeePerGas, '150');
        });
        it('should build txn with eip1559 for celo', async function () {
            const txBuilder = (0, getBuilder_1.getBuilder)('celo');
            txBuilder.fee({
                fee: '1000000000',
                gasLimit: '12100000',
                eip1559: {
                    maxFeePerGas: '7593123',
                    maxPriorityFeePerGas: '150',
                },
            });
            txBuilder.counter(2);
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.contract('0x8f977e912ef500548a0c3be6ddde9899f1199b81');
            txBuilder
                .transfer()
                .coin('celo')
                .amount('1000000000')
                .to('0x19645032c7f1533395d44a629462e751084d3e4c')
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            txBuilder.sign({ key: testData.PRIVATE_KEY });
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.gasLimit, '12100000');
            should_1.default.equal(txJson._type, 'EIP1559');
            should_1.default.equal(txJson.maxFeePerGas, '7593123');
            should_1.default.equal(txJson.maxPriorityFeePerGas, '150');
        });
        it('should build txn with eip1559 for celo token', async function () {
            const txBuilder = (0, getBuilder_1.getBuilder)('cusd');
            txBuilder.fee({
                fee: '1000000000',
                gasLimit: '12100000',
                eip1559: {
                    maxFeePerGas: '7593123',
                    maxPriorityFeePerGas: '150',
                },
            });
            txBuilder.counter(2);
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.contract('0x8f977e912ef500548a0c3be6ddde9899f1199b81');
            txBuilder
                .transfer()
                .coin('cusd')
                .amount('1000000000')
                .to('0x19645032c7f1533395d44a629462e751084d3e4c')
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            txBuilder.sign({ key: testData.PRIVATE_KEY });
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.gasLimit, '12100000');
            should_1.default.equal(txJson._type, 'EIP1559');
            should_1.default.equal(txJson.maxFeePerGas, '7593123');
            should_1.default.equal(txJson.maxPriorityFeePerGas, '150');
        });
        it('should decode token txn', async function () {
            const txBuilder = (0, getBuilder_1.getBuilder)('cusd');
            txBuilder.from(testData.PROD_SEND_TOKEN_BROADCAST);
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.gasLimit, '12100000');
            should_1.default.equal(txJson._type, 'EIP1559');
            should_1.default.equal(txJson.maxFeePerGas, '7593123');
            should_1.default.equal(txJson.maxPriorityFeePerGas, '150');
        });
        it('should build legacy txn for celo', async function () {
            const txBuilder = (0, getBuilder_1.getBuilder)('celo');
            txBuilder.fee({
                fee: '1000000000',
                gasLimit: '12100000',
            });
            txBuilder.counter(2);
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.contract('0x8f977e912ef500548a0c3be6ddde9899f1199b81');
            txBuilder
                .transfer()
                .coin('celo')
                .amount('1000000000')
                .to('0x19645032c7f1533395d44a629462e751084d3e4c')
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.gasLimit, '12100000');
            should_1.default.equal(txJson._type, 'Legacy');
            should_1.default.equal(txJson.v, '0x0149fb');
        });
        it('a send token transaction without final v', async () => {
            const recipient = '0x19645032c7f1533395d44a629462e751084d3e4c';
            const contractAddress = '0x8f977e912ef500548a0c3be6ddde9899f1199b81';
            const amount = '1000000000';
            initTxBuilder();
            txBuilder.contract(contractAddress);
            txBuilder
                .transfer()
                .coin('tcusd')
                .amount(amount)
                .to(recipient)
                .expirationTime(1590066728)
                .contractSequenceId(5)
                .key(key);
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.equal(txJson.v, '0x015408bb');
        });
    });
    describe('should fail to build', async () => {
        it('a send token transaction without fee', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.counter(1);
            txBuilder.contract(testData.CONTRACT_TOKEN_CUSD_ADDRESS);
            await txBuilder.build().should.be.rejectedWith('Invalid transaction: missing fee');
        });
        it('a send token transaction without transfer information', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.fee({
                fee: '10000000000',
                gasLimit: '2000000',
            });
            txBuilder.counter(1);
            txBuilder.contract(testData.CONTRACT_TOKEN_CUSD_ADDRESS);
            await txBuilder.build().should.be.rejectedWith('Missing transfer information');
        });
        it('a send token transaction without nonce', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.fee({
                fee: '10000000000',
                gasLimit: '2000000',
            });
            txBuilder.counter(1);
            await txBuilder.build().should.be.rejectedWith('Invalid transaction: missing contract address');
        });
        it('a send token transaction with wrong transaction type', async () => {
            initTxBuilder();
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.contract('0x8f977e912ef500548a0c3be6ddde9899f1199b81');
            assert_1.default.throws(() => {
                txBuilder.transfer();
            }, 'Error: Token transfers can only be set for send token transactions');
        });
        it('a send token transaction without token information', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('tcelo');
            txBuilder.type(sdk_core_1.TransactionType.Send);
            txBuilder.fee({
                fee: '10000000000',
                gasLimit: '2000000',
            });
            txBuilder.counter(1);
            txBuilder.contract(testData.CONTRACT_TOKEN_CUSD_ADDRESS);
            await txBuilder.build().should.be.rejectedWith('Missing transfer information');
        });
    });
});
