// @substrate/txwrapper-core/lib/core/metadata/createMetadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMetadata = exports.createMetadataUnmemoized = void 0;
const types_1 = require("@polkadot/types");
const MetadataVersioned_1 = require("@polkadot/types/metadata/MetadataVersioned");
const memoizee_1 = __importDefault(require("memoizee"));
const util_1 = require("../util");
const toSpecifiedCallsOnlyV14_1 = require("./toSpecifiedCallsOnlyV14");
/**
 * From a metadata hex string (for example returned by RPC), create a Metadata
 * class instance. Metadata decoding is expensive, so this function is
 * memoized.
 *
 * @ignore
 * @param registry - The registry of the metadata.
 * @param metadata - The metadata as hex string.
 * @param asCallsOnlyArg - Option to decreases the metadata to calls only
 */
function createMetadataUnmemoized(registry, metadataRpc, asCallsOnlyArg = false, asSpecifiedCallsOnlyV14) {
    const metadata = new types_1.Metadata(registry, metadataRpc);
    if (asSpecifiedCallsOnlyV14 && asSpecifiedCallsOnlyV14.length > 0) {
        return new MetadataVersioned_1.MetadataVersioned(registry, {
            magicNumber: metadata.magicNumber,
            metadata: registry.createTypeUnsafe('MetadataAll', [
                (0, toSpecifiedCallsOnlyV14_1.toSpecifiedCallsOnlyV14)(registry, metadata.asV14, asSpecifiedCallsOnlyV14),
                14,
            ]),
        });
    }
    return asCallsOnlyArg ? metadata.asCallsOnly : metadata;
}
exports.createMetadataUnmemoized = createMetadataUnmemoized;
/**
 * From a metadata hex string (for example returned by RPC), create a Metadata
 * class instance. Metadata decoding is expensive, so this function is
 * memoized.
 *
 * @ignore
 * @param registry - The registry of the metadata.
 * @param metadata - The metadata as hex string.
 * @param asCallsOnlyArg - Option to decreases the metadata to calls only
 */
exports.createMetadata = (0, memoizee_1.default)(createMetadataUnmemoized, {
    length: 4,
    max: !util_1.isBrowser &&
        typeof process !== 'undefined' &&
        typeof ((_a = process.env) === null || _a === void 0 ? void 0 : _a.TXWRAPPER_METADATA_CACHE_MAX) !== 'undefined'
        ? parseInt(process.env.TXWRAPPER_METADATA_CACHE_MAX)
        : undefined,
    maxAge: !util_1.isBrowser &&
        typeof process !== 'undefined' &&
        typeof ((_b = process.env) === null || _b === void 0 ? void 0 : _b.TXWRAPPER_METADATA_CACHE_MAX_AGE) !== 'undefined'
        ? parseInt(process.env.TXWRAPPER_METADATA_CACHE_MAX_AGE)
        : undefined,
});
