// @bitgo/sdk-core/dist/src/bitgo/coinFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalCoinFactory = exports.CoinFactory = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const errors_1 = require("./errors");
class CoinFactory {
    constructor(coinMap = statics_1.coins) {
        this.coinConstructors = new Map();
        this.coinMap = coinMap;
    }
    /**
     * @param name Name of coin or address
     * @returns {(Object|undefined)}
     */
    getCoinConstructor(name) {
        if (this.coinConstructors === undefined) {
            this.coinConstructors = new Map();
        }
        return this.coinConstructors.get(name);
    }
    /**
     * @param name Name of coin or address
     * @param coin Coin plugin's constructor
     * @throws Error
     */
    register(name, coin) {
        this.coinConstructors.set(name, coin);
    }
    /**
     * Checks if the coin is present in both coin map and constructor map
     * @param name Name of coin
     * @returns {boolean}
     */
    hasCoin(name) {
        return this.coinMap.has(name) && this.getCoinConstructor(name) !== undefined;
    }
    /**
     * Registers a token in the coin map and the constructor map.
     * @param staticsCoin The static coin definition from BitGo Statics
     * @param coinConstructor The constructor for the coin plugin
     * @throws Error
     */
    registerToken(staticsCoin, coinConstructor) {
        if (!(this.coinMap.has(staticsCoin.name) ||
            this.coinMap.has(staticsCoin.id) ||
            (staticsCoin.alias && this.coinMap.has(staticsCoin.alias)))) {
            this.coinMap.addCoin(staticsCoin);
        }
        this.register(staticsCoin.name, coinConstructor);
    }
    /**
     * @param bitgo Instance of BitGo
     * @param name Name of coin or address
     * @throws CoinNotDefinedError
     * @throws UnsupportedCoinError
     */
    getInstance(bitgo, name) {
        let staticsCoin;
        try {
            staticsCoin = this.coinMap.get(name);
        }
        catch (e) {
            if (!(e instanceof statics_1.CoinNotDefinedError)) {
                throw e;
            }
        }
        const constructor = this.getCoinConstructor(name);
        if (constructor) {
            return constructor(bitgo, staticsCoin);
        }
        throw new errors_1.UnsupportedCoinError(name);
    }
}
exports.CoinFactory = CoinFactory;
exports.GlobalCoinFactory = new CoinFactory();
