// @bitgo/sdk-core/dist/src/bitgo/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxIntentMismatchApprovalError = exports.TxIntentMismatchContractError = exports.TxIntentMismatchRecipientError = exports.TxIntentMismatchError = exports.ApiResponseError = exports.NeedUserSignupError = exports.IncorrectPasswordError = exports.MissingEncryptedKeychainError = exports.InvalidTransactionError = exports.InvalidKeyPathError = exports.ErrorNoInputToRecover = exports.StellarFederationUserNotFoundError = exports.EthereumLibraryUnavailableError = exports.AddressGenerationError = exports.KeyRecoveryServiceError = exports.InvalidPaymentIdError = exports.InvalidMemoIdError = exports.BlockExplorerUnavailable = exports.MethodNotImplementedError = exports.WalletRecoveryUnsupported = exports.InvalidAddressDerivationPropertyError = exports.UnexpectedAddressError = exports.InvalidAddressVerificationObjectPropertyError = exports.InvalidAddressError = exports.UnsupportedAddressTypeError = exports.P2trMusig2UnsupportedError = exports.P2trUnsupportedError = exports.P2wshUnsupportedError = exports.P2shP2wshUnsupportedError = exports.AddressTypeChainMismatchError = exports.UnsupportedCoinError = exports.NodeEnvironmentError = exports.TlsConfigurationError = exports.BitGoJsError = void 0;
// Descriptive error types for common issues which may arise
// during the operation of BitGoJS or BitGoExpress
const bitgojsError_1 = require("../bitgojsError");
Object.defineProperty(exports, "BitGoJsError", { enumerable: true, get: function () { return bitgojsError_1.BitGoJsError; } });
class TlsConfigurationError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'TLS is configuration is invalid');
    }
}
exports.TlsConfigurationError = TlsConfigurationError;
class NodeEnvironmentError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'NODE_ENV is invalid for the current bitgo environment');
    }
}
exports.NodeEnvironmentError = NodeEnvironmentError;
class UnsupportedCoinError extends bitgojsError_1.BitGoJsError {
    constructor(coin) {
        super(`Coin or token type ${coin} not supported or not compiled. Please be sure that you are using the latest version of BitGoJS. If using @bitgo/sdk-api, please confirm you have registered ${coin} first.`);
    }
}
exports.UnsupportedCoinError = UnsupportedCoinError;
class AddressTypeChainMismatchError extends bitgojsError_1.BitGoJsError {
    constructor(addressType, chain) {
        super(`address type ${addressType} does not correspond to chain ${chain}`);
    }
}
exports.AddressTypeChainMismatchError = AddressTypeChainMismatchError;
class P2shP2wshUnsupportedError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'p2shP2wsh not supported by this coin');
    }
}
exports.P2shP2wshUnsupportedError = P2shP2wshUnsupportedError;
class P2wshUnsupportedError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'p2wsh not supported by this coin');
    }
}
exports.P2wshUnsupportedError = P2wshUnsupportedError;
class P2trUnsupportedError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'p2tr not supported by this coin');
    }
}
exports.P2trUnsupportedError = P2trUnsupportedError;
class P2trMusig2UnsupportedError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'p2trMusig2 not supported by this coin');
    }
}
exports.P2trMusig2UnsupportedError = P2trMusig2UnsupportedError;
class UnsupportedAddressTypeError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'invalid address type');
    }
}
exports.UnsupportedAddressTypeError = UnsupportedAddressTypeError;
class InvalidAddressError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'invalid address');
    }
}
exports.InvalidAddressError = InvalidAddressError;
class InvalidAddressVerificationObjectPropertyError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'address validation failure');
    }
}
exports.InvalidAddressVerificationObjectPropertyError = InvalidAddressVerificationObjectPropertyError;
class UnexpectedAddressError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'address validation failure');
    }
}
exports.UnexpectedAddressError = UnexpectedAddressError;
class InvalidAddressDerivationPropertyError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'address chain and/or index are invalid');
    }
}
exports.InvalidAddressDerivationPropertyError = InvalidAddressDerivationPropertyError;
class WalletRecoveryUnsupported extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'wallet recovery is not supported by this coin');
    }
}
exports.WalletRecoveryUnsupported = WalletRecoveryUnsupported;
class MethodNotImplementedError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'method not implemented');
    }
}
exports.MethodNotImplementedError = MethodNotImplementedError;
class BlockExplorerUnavailable extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'third-party blockexplorer not responding');
    }
}
exports.BlockExplorerUnavailable = BlockExplorerUnavailable;
class InvalidMemoIdError extends InvalidAddressError {
    constructor(message) {
        super(message || 'invalid memo id');
    }
}
exports.InvalidMemoIdError = InvalidMemoIdError;
class InvalidPaymentIdError extends InvalidAddressError {
    constructor(message) {
        super(message || 'invalid payment id');
    }
}
exports.InvalidPaymentIdError = InvalidPaymentIdError;
class KeyRecoveryServiceError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'key recovery service encountered an error');
    }
}
exports.KeyRecoveryServiceError = KeyRecoveryServiceError;
class AddressGenerationError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'address generation failed');
    }
}
exports.AddressGenerationError = AddressGenerationError;
class EthereumLibraryUnavailableError extends bitgojsError_1.BitGoJsError {
    constructor(packageName) {
        super(`Ethereum library required for operation is not available. Please install "${packageName}".`);
    }
}
exports.EthereumLibraryUnavailableError = EthereumLibraryUnavailableError;
class StellarFederationUserNotFoundError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'account not found');
    }
}
exports.StellarFederationUserNotFoundError = StellarFederationUserNotFoundError;
class ErrorNoInputToRecover extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'No input to recover - aborting!');
    }
}
exports.ErrorNoInputToRecover = ErrorNoInputToRecover;
class InvalidKeyPathError extends bitgojsError_1.BitGoJsError {
    constructor(keyPath) {
        super(`invalid keypath: ${keyPath}`);
    }
}
exports.InvalidKeyPathError = InvalidKeyPathError;
class InvalidTransactionError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'Invalid transaction');
    }
}
exports.InvalidTransactionError = InvalidTransactionError;
class MissingEncryptedKeychainError extends Error {
    constructor(message) {
        super(message || 'No encrypted keychains on this wallet.');
    }
}
exports.MissingEncryptedKeychainError = MissingEncryptedKeychainError;
class IncorrectPasswordError extends Error {
    constructor(message) {
        super(message || 'Incorrect password');
    }
}
exports.IncorrectPasswordError = IncorrectPasswordError;
class NeedUserSignupError extends bitgojsError_1.BitGoJsError {
    constructor(message) {
        super(message || 'User signup is required');
    }
}
exports.NeedUserSignupError = NeedUserSignupError;
class ApiResponseError extends bitgojsError_1.BitGoJsError {
    constructor(message, status, result, invalidToken, needsOTP) {
        super(message);
        this.message = message;
        this.status = status;
        this.result = result;
        this.invalidToken = invalidToken;
        this.needsOTP = needsOTP;
    }
}
exports.ApiResponseError = ApiResponseError;
/**
 * Base error class for transaction intent mismatch detection
 *
 * This error is thrown when a transaction does not match the user's original intent,
 * indicating potential security issues or malicious modifications.
 *
 * @class TxIntentMismatchError
 * @extends {BitGoJsError}
 * @property {string | IRequestTracer | undefined} id - Transaction ID or request tracer for tracking
 * @property {TransactionParams[]} txParams - Array of transaction parameters that were analyzed
 * @property {string | undefined} txHex - The raw transaction in hexadecimal format
 * @property {string | undefined} txExplanation - Stringified transaction explanation
 */
class TxIntentMismatchError extends bitgojsError_1.BitGoJsError {
    /**
     * Creates an instance of TxIntentMismatchError
     *
     * @param {string} message - Error message describing the intent mismatch
     * @param {string | IRequestTracer | undefined} id - Transaction ID or request tracer
     * @param {TransactionParams[]} txParams - Transaction parameters that were analyzed
     * @param {string | undefined} txHex - Raw transaction hex string
     * @param {unknown} txExplanation - Transaction explanation
     */
    constructor(message, id, txParams, txHex, txExplanation) {
        super(message);
        this.id = id;
        this.txParams = txParams;
        this.txHex = txHex;
        this.txExplanation = txExplanation ? TxIntentMismatchError.safeStringify(txExplanation) : undefined;
    }
    /**
     * Safely stringify a value with BigInt support
     * @param value - Value to stringify
     * @returns JSON string with BigInts converted to strings
     */
    static safeStringify(value) {
        return JSON.stringify(value, (_, v) => (typeof v === 'bigint' ? v.toString() : v), 2);
    }
    /**
     * Try to get transaction explanation from a coin's explainTransaction method.
     *
     * @param coin - Coin instance implementing IBaseCoin
     * @param txPrebuild - Transaction prebuild containing txHex and txInfo
     * @returns Transaction explanation object or undefined
     */
    static async tryGetTxExplanation(coin, txPrebuild) {
        if (!txPrebuild.txHex) {
            return undefined;
        }
        try {
            return await coin.explainTransaction({
                txHex: txPrebuild.txHex,
                txInfo: txPrebuild.txInfo,
            });
        }
        catch (e) {
            return {
                error: 'Failed to parse transaction explanation',
                txHex: txPrebuild.txHex,
                details: e instanceof Error ? e.message : String(e),
            };
        }
    }
}
exports.TxIntentMismatchError = TxIntentMismatchError;
/**
 * Error thrown when transaction recipients don't match the user's intent
 *
 * This error occurs when the transaction contains recipients or amounts that differ
 * from what the user originally intended to send.
 *
 * @class TxIntentMismatchRecipientError
 * @extends {TxIntentMismatchError}
 * @property {MismatchedRecipient[]} mismatchedRecipients - Array of recipients that don't match user intent
 */
class TxIntentMismatchRecipientError extends TxIntentMismatchError {
    /**
     * Creates an instance of TxIntentMismatchRecipientError
     *
     * @param {string} message - Error message describing the recipient intent mismatch
     * @param {string | IRequestTracer | undefined} id - Transaction ID or request tracer
     * @param {TransactionParams[]} txParams - Transaction parameters that were analyzed
     * @param {string | undefined} txHex - Raw transaction hex string
     * @param {MismatchedRecipient[]} mismatchedRecipients - Array of recipients that don't match user intent
     * @param {unknown} txExplanation - Transaction explanation
     */
    constructor(message, id, txParams, txHex, mismatchedRecipients, txExplanation) {
        super(message, id, txParams, txHex, txExplanation);
        this.mismatchedRecipients = mismatchedRecipients;
    }
}
exports.TxIntentMismatchRecipientError = TxIntentMismatchRecipientError;
/**
 * Error thrown when contract interaction doesn't match the user's intent
 *
 * This error occurs when a transaction interacts with a smart contract but the
 * contract call data or method doesn't match what the user intended.
 *
 * @class TxIntentMismatchContractError
 * @extends {TxIntentMismatchError}
 * @property {ContractDataPayload} mismatchedDataPayload - The contract interaction data that doesn't match user intent
 */
class TxIntentMismatchContractError extends TxIntentMismatchError {
    /**
     * Creates an instance of TxIntentMismatchContractError
     *
     * @param {string} message - Error message describing the contract intent mismatch
     * @param {string | IRequestTracer | undefined} id - Transaction ID or request tracer
     * @param {TransactionParams[]} txParams - Transaction parameters that were analyzed
     * @param {string | undefined} txHex - Raw transaction hex string
     * @param {ContractDataPayload} mismatchedDataPayload - The contract interaction data that doesn't match user intent
     * @param {unknown} txExplanation - Transaction explanation
     */
    constructor(message, id, txParams, txHex, mismatchedDataPayload, txExplanation) {
        super(message, id, txParams, txHex, txExplanation);
        this.mismatchedDataPayload = mismatchedDataPayload;
    }
}
exports.TxIntentMismatchContractError = TxIntentMismatchContractError;
/**
 * Error thrown when token approval doesn't match the user's intent
 *
 * This error occurs when a transaction contains a token approval that the user
 * did not intend to authorize, potentially indicating malicious activity.
 *
 * @class TxIntentMismatchApprovalError
 * @extends {TxIntentMismatchError}
 * @property {TokenApproval} tokenApproval - Details of the token approval that doesn't match user intent
 */
class TxIntentMismatchApprovalError extends TxIntentMismatchError {
    /**
     * Creates an instance of TxIntentMismatchApprovalError
     *
     * @param {string} message - Error message describing the approval intent mismatch
     * @param {string | IRequestTracer | undefined} id - Transaction ID or request tracer
     * @param {TransactionParams[]} txParams - Transaction parameters that were analyzed
     * @param {string | undefined} txHex - Raw transaction hex string
     * @param {TokenApproval} tokenApproval - Details of the token approval that doesn't match user intent
     * @param {unknown} txExplanation - Transaction explanation
     */
    constructor(message, id, txParams, txHex, tokenApproval, txExplanation) {
        super(message, id, txParams, txHex, txExplanation);
        this.tokenApproval = tokenApproval;
    }
}
exports.TxIntentMismatchApprovalError = TxIntentMismatchApprovalError;
