// @bitgo/sdk-core/dist/src/bitgo/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tss = exports.bitcoinUtil = exports.sendSignatureShare = exports.WalletProofs = exports.bitcoin = exports.walletUtil = void 0;
const bitcoinUtil = __importStar(require("./bitcoin"));
exports.bitcoinUtil = bitcoinUtil;
const tss = __importStar(require("./tss"));
exports.tss = tss;
exports.walletUtil = __importStar(require("./walletUtil"));
__exportStar(require("./baseCoin"), exports);
__exportStar(require("./bip32util"), exports);
__exportStar(require("./bitcoin"), exports);
__exportStar(require("./bitgoBase"), exports);
__exportStar(require("./config"), exports);
__exportStar(require("./coinFactory"), exports);
__exportStar(require("./ecdh"), exports);
__exportStar(require("./enterprise"), exports);
__exportStar(require("./environments"), exports);
__exportStar(require("./errors"), exports);
__exportStar(require("./inscriptionBuilder"), exports);
__exportStar(require("./internal"), exports);
__exportStar(require("./keychain"), exports);
exports.bitcoin = __importStar(require("./legacyBitcoin"));
__exportStar(require("./market"), exports);
__exportStar(require("./pendingApproval"), exports);
var proofs_1 = require("./proofs");
Object.defineProperty(exports, "WalletProofs", { enumerable: true, get: function () { return proofs_1.WalletProofs; } });
__exportStar(require("./recovery"), exports);
__exportStar(require("./staking"), exports);
__exportStar(require("./trading"), exports);
__exportStar(require("./tss"), exports);
var tss_1 = require("./tss");
Object.defineProperty(exports, "sendSignatureShare", { enumerable: true, get: function () { return tss_1.sendSignatureShare; } });
__exportStar(require("./types"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./wallet"), exports);
__exportStar(require("./webhook"), exports);
