// @bitgo/sdk-core/dist/src/bitgo/legacyBitcoin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.address = exports.networks = exports.HDNode = void 0;
exports.makeRandomKey = makeRandomKey;
exports.hdPath = hdPath;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const api_1 = require("../api");
const bitcoinUtil = __importStar(require("./bitcoin"));
function createECPairCompat(ecPair) {
    return Object.assign(ecPair, {
        getPublicKeyBuffer() {
            return ecPair.publicKey;
        },
    });
}
function makeRandomKey() {
    return createECPairCompat(bitcoinUtil.makeRandomKey());
}
/**
 * Implementation of legacy "HDNode" class as used by certain components
 */
class HDNode {
    constructor(bip32) {
        this.bip32 = bip32;
    }
    get publicKey() {
        return this.bip32.publicKey;
    }
    get privateKey() {
        return this.bip32.privateKey;
    }
    static fromSeedBuffer(buf) {
        return new HDNode(utxo_lib_1.bip32.fromSeed(buf));
    }
    static fromBase58(str) {
        return new HDNode(utxo_lib_1.bip32.fromBase58(str));
    }
    isNeutered() {
        return this.bip32.isNeutered();
    }
    neutered() {
        return new HDNode(this.bip32.neutered());
    }
    toBase58() {
        return this.bip32.toBase58();
    }
    derivePath(p) {
        return new HDNode(this.bip32.derivePath((0, api_1.sanitizeLegacyPath)(p)));
    }
}
exports.HDNode = HDNode;
function hdPath(hdNode) {
    return {
        derive(path) {
            return hdNode.derivePath(path);
        },
        deriveKey(path) {
            const node = hdNode.derivePath(path);
            if (node.privateKey) {
                return createECPairCompat(utxolib.ECPair.fromPrivateKey(node.privateKey));
            }
            else {
                return createECPairCompat(utxolib.ECPair.fromPublicKey(node.publicKey));
            }
        },
    };
}
exports.networks = utxolib.networks;
exports.address = {
    fromBase58Check(addr) {
        return utxolib.address.fromBase58Check(addr, utxolib.networks.bitcoin);
    },
    toBase58Check(hash, version) {
        return utxolib.address.toBase58Check(hash, version, utxolib.networks.bitcoin);
    },
};
