// @bitgo/sdk-coin-cosmos/dist/src/cosmosSharedCoin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSharedCoin = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const lib_1 = require("./lib");
/**
 * Shared Cosmos coin implementation that uses configuration from statics
 * instead of requiring individual coin modules
 */
class CosmosSharedCoin extends abstract_cosmos_1.CosmosCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
        // Get the network configuration from statics
        const network = staticsCoin.network;
        if (!network || !network.addressPrefix || !network.validatorPrefix) {
            throw new Error('Invalid network configuration: missing required Cosmos network parameters');
        }
        this._network = network;
    }
    static createInstance(bitgo, staticsCoin) {
        return new CosmosSharedCoin(bitgo, staticsCoin);
    }
    /** @inheritDoc **/
    getBaseFactor() {
        return Math.pow(10, this._staticsCoin.decimalPlaces);
    }
    /** @inheritDoc **/
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(this._staticsCoin);
    }
    /** @inheritDoc **/
    isValidAddress(address) {
        const utils = new lib_1.Utils(this._network);
        return utils.isValidAddress(address) || utils.isValidValidatorAddress(address);
    }
    /** @inheritDoc **/
    getDenomination() {
        return this._network.denom;
    }
    /** @inheritDoc **/
    getGasAmountDetails() {
        return {
            gasAmount: this._network.gasAmount,
            gasLimit: this._network.gasLimit,
        };
    }
    /**
     * Get the network configuration for this coin
     * @returns {CosmosNetwork} The network configuration
     */
    getNetwork() {
        return this._network;
    }
    /** @inheritDoc **/
    getKeyPair(publicKey) {
        return new lib_1.KeyPair({ pub: publicKey }, this._staticsCoin);
    }
    /** @inheritDoc **/
    getAddressFromPublicKey(pubKey) {
        const keyPair = new lib_1.KeyPair({ pub: pubKey }, this._staticsCoin);
        return keyPair.getAddress();
    }
    /** @inheritDoc **/
    getPublicNodeUrl() {
        const family = this.getFamily();
        const env = this.bitgo.getEnv();
        const cosmosConfig = sdk_core_1.common.Environments[env]?.cosmos;
        (0, assert_1.default)(cosmosConfig?.[family]?.nodeUrl, `env config is missing for ${family} in ${env}`);
        return cosmosConfig[family].nodeUrl;
    }
}
exports.CosmosSharedCoin = CosmosSharedCoin;
