// @bitgo/abstract-utxo/dist/cjs/src/recovery/backupKeyRecovery.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V2 = exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V1_SWEEP = exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V1 = void 0;
exports.backupKeyRecoveryWithWalletUnspents = backupKeyRecoveryWithWalletUnspents;
exports.formatBackupKeyRecoveryResult = formatBackupKeyRecoveryResult;
exports.backupKeyRecovery = backupKeyRecovery;
exports.v1BackupKeyRecovery = v1BackupKeyRecovery;
exports.v1Sweep = v1Sweep;
const lodash_1 = __importDefault(require("lodash"));
const sdk_core_1 = require("@bitgo/sdk-core");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const signTransaction_1 = require("../transaction/fixedScript/signTransaction");
const address_1 = require("../address");
const decode_1 = require("../transaction/decode");
const replayProtection_1 = require("../transaction/fixedScript/replayProtection");
const names_1 = require("../names");
const unspent_1 = require("../unspent");
const RecoveryProvider_1 = require("./RecoveryProvider");
const mempoolApi_1 = require("./mempoolApi");
const coingeckoApi_1 = require("./coingeckoApi");
const psbt_1 = require("./psbt");
// V1 only deals with BTC. 50 sat/vbyte is very arbitrary.
exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V1 = 50;
// FIXME(BTC-2691): it is unclear why sweeps have a different default than regular recovery. 100 sat/vbyte is extremely high.
exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V1_SWEEP = 100;
// FIXME(BTC-2691): it makes little sense to have a single default for every coin.
exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V2 = 50;
/**
 * Get the current market price from a third party to be used for recovery
 * This function is only intended for non-bitgo recovery transactions, when it is necessary
 * to calculate the rough fee needed to pay to Keyternal. We are okay with approximating,
 * because the resulting price of this function only has less than 1 dollar influence on the
 * fee that needs to be paid to Keyternal.
 *
 * See calculateFeeAmount function:  return Math.round(feeAmountUsd / currentPrice * self.getBaseFactor());
 *
 * This end function should not be used as an accurate endpoint, since some coins' prices are missing from the provider
 */
async function getRecoveryMarketPrice(coin) {
    return await new coingeckoApi_1.CoingeckoApi().getUSDPrice(coin.getFamily());
}
/**
 * Calculates the amount (in base units) to pay a KRS provider when building a recovery transaction
 * @param coin
 * @param params
 * @param params.provider {String} the KRS provider that holds the backup key
 * @returns {*}
 */
async function calculateFeeAmount(coin, params) {
    const krsProvider = sdk_core_1.krsProviders[params.provider];
    if (krsProvider === undefined) {
        throw new Error(`no fee structure specified for provider ${params.provider}`);
    }
    if (krsProvider.feeType === 'flatUsd') {
        const feeAmountUsd = krsProvider.feeAmount;
        const currentPrice = await getRecoveryMarketPrice(coin);
        return Math.round((feeAmountUsd / currentPrice) * coin.getBaseFactor());
    }
    else {
        // we can add more fee structures here as needed for different providers, such as percentage of recovery amount
        throw new Error('Fee structure not implemented');
    }
}
/**
 * Generate an address and format it for API queries
 * @param coin - The coin instance
 * @param coinName - The coin name
 * @param walletKeys - The wallet keys
 * @param chain - The chain code
 * @param addrIndex - The address index
 * @returns The formatted address (with cashaddr prefix stripped for BCH/BCHA)
 */
function getFormattedAddress(coin, coinName, walletKeys, chain, addrIndex) {
    const format = coin.getChain() === 'bch' || coin.getChain() === 'bcha' ? 'cashaddr' : undefined;
    const address = (0, address_1.generateAddressWithChainAndIndex)(coinName, walletKeys, chain, addrIndex, format);
    // Blockchair uses cashaddr format when querying the API for address information. Strip the prefix for BCH/BCHA.
    return format === 'cashaddr' ? address.split(':')[1] : address;
}
function hasWitnessData(scriptType) {
    return scriptType !== 'p2sh';
}
async function queryBlockchainUnspentsPath(coin, params, walletKeys, chain) {
    const scriptType = wasm_utxo_1.fixedScriptWallet.ChainCode.scriptType(chain);
    const fetchPrevTx = !hasWitnessData(scriptType) && (0, names_1.getMainnetCoinName)(coin.name) !== 'zec';
    const recoveryProvider = params.recoveryProvider ?? (0, RecoveryProvider_1.forCoin)(coin.getChain(), params.apiKey);
    const MAX_SEQUENTIAL_ADDRESSES_WITHOUT_TXS = params.scan || 20;
    let numSequentialAddressesWithoutTxs = 0;
    const prevTxCache = new Map();
    async function getPrevTx(txid) {
        let prevTxHex = prevTxCache.get(txid);
        if (!prevTxHex) {
            prevTxHex = await recoveryProvider.getTransactionHex(txid);
            prevTxCache.set(txid, prevTxHex);
        }
        return prevTxHex;
    }
    async function gatherUnspents(addrIndex) {
        const formattedAddress = getFormattedAddress(coin, coin.name, walletKeys, chain, addrIndex);
        const addrInfo = await recoveryProvider.getAddressInfo(formattedAddress);
        // we use txCount here because it implies usage - having tx'es means the addr was generated and used
        if (addrInfo.txCount === 0) {
            numSequentialAddressesWithoutTxs++;
        }
        else {
            numSequentialAddressesWithoutTxs = 0;
            if (addrInfo.balance > 0) {
                console.log(`Found an address with balance: ${formattedAddress} with balance ${addrInfo.balance}`);
                const addressUnspents = await recoveryProvider.getUnspentsForAddresses([formattedAddress]);
                const processedUnspents = await Promise.all(addressUnspents.map(async (u) => {
                    const { txid, vout } = (0, unspent_1.parseOutputId)(u.id);
                    let val = BigInt(u.value);
                    if (coin.amountType === 'bigint') {
                        // blockchair returns the number with the correct precision, but in number format
                        // json parse won't parse it correctly, so we requery the txid for the tx hex to decode here
                        if (!Number.isSafeInteger(u.value)) {
                            const txHex = await getPrevTx(txid);
                            const tx = coin.createTransactionFromHex(txHex);
                            val = tx.outs[vout].value;
                        }
                    }
                    // the api may return cashaddr's instead of legacy for BCH and BCHA
                    // downstream processes's only expect legacy addresses
                    u = { ...u, address: coin.canonicalAddress(u.address) };
                    return {
                        ...u,
                        value: val,
                        chain: chain,
                        index: addrIndex,
                        prevTx: fetchPrevTx ? Buffer.from(await getPrevTx(txid), 'hex') : undefined,
                    };
                }));
                walletUnspents.push(...processedUnspents);
            }
        }
        if (numSequentialAddressesWithoutTxs >= MAX_SEQUENTIAL_ADDRESSES_WITHOUT_TXS) {
            // stop searching for addresses with unspents in them, we've found ${MAX_SEQUENTIAL_ADDRESSES_WITHOUT_TXS} in a row with none
            // we are done
            return;
        }
        return gatherUnspents(addrIndex + 1);
    }
    // get unspents for these addresses
    const walletUnspents = [];
    // This will populate walletAddresses
    await gatherUnspents(0);
    if (walletUnspents.length === 0) {
        // Couldn't find any addresses with funds
        return [];
    }
    return walletUnspents;
}
async function getRecoveryFeePerBytes(coin, { defaultValue }) {
    if (coin.getFamily() === 'doge') {
        // https://github.com/dogecoin/dogecoin/blob/master/doc/fee-recommendation.md
        // 0.01 DOGE per KB
        return 1000;
    }
    try {
        return await mempoolApi_1.MempoolApi.forCoin(coin.getChain()).getRecoveryFeePerBytes();
    }
    catch (e) {
        console.dir(e);
        return defaultValue;
    }
}
function hasPrivateKey(key) {
    return key.privateKey !== undefined;
}
/**
 * Builds a funds recovery PSBT without BitGo, using provided unspents.
 *
 * This is the core transaction building logic, separated from unspent gathering
 * and output formatting. Returns a PSBT at the appropriate signing stage.
 *
 * Signing behavior is determined by the keys:
 * - If user key has no private key: unsigned PSBT
 * - If user key has private key but backup doesn't: half-signed PSBT (user signature only)
 * - If both user and backup keys have private keys: fully signed PSBT (not finalized)
 *
 * @param coinName - The coin name for the PSBT
 * @param params - Recovery parameters with pre-derived keys
 * @param unspents - The wallet unspents to recover (must be non-empty)
 * @returns The PSBT at the appropriate signing stage (never finalized)
 */
function backupKeyRecoveryWithWalletUnspents(coinName, params, unspents) {
    const { walletKeys, keys, recoveryDestination, feeRateSatVB, krsFee, krsFeeAddress } = params;
    const totalInputAmount = (0, unspent_1.unspentSum)(unspents);
    if (totalInputAmount <= BigInt(0)) {
        throw new sdk_core_1.ErrorNoInputToRecover();
    }
    let psbt = (0, psbt_1.createBackupKeyRecoveryPsbt)(coinName, walletKeys, unspents, {
        feeRateSatVB: feeRateSatVB,
        recoveryDestination: recoveryDestination,
        keyRecoveryServiceFee: krsFee ?? BigInt(0),
        keyRecoveryServiceFeeAddress: krsFeeAddress,
    });
    const userHasPrivateKey = hasPrivateKey(keys[0]);
    const backupHasPrivateKey = hasPrivateKey(keys[1]);
    if (!userHasPrivateKey) {
        // Unsigned sweep - return unsigned PSBT
        return psbt;
    }
    const replayProtection = { publicKeys: (0, replayProtection_1.getReplayProtectionPubkeys)(coinName) };
    // Sign with user key
    psbt = (0, signTransaction_1.signAndVerifyPsbt)(psbt, keys[0], walletKeys, replayProtection);
    if (backupHasPrivateKey) {
        // Full recovery - sign with backup key too
        psbt = (0, signTransaction_1.signAndVerifyPsbt)(psbt, keys[1], walletKeys, replayProtection);
    }
    // Return PSBT (not finalized - let caller decide how to format)
    return psbt;
}
/**
 * Formats a backup key recovery PSBT into the appropriate response format.
 *
 * Output format depends on signing state and KRS provider:
 * - Unsigned sweep: FormattedOfflineVaultTxInfo with PSBT hex
 * - KRS keyternal: BackupKeyRecoveryTransansaction with legacy half-signed tx hex
 * - KRS other: BackupKeyRecoveryTransansaction with PSBT hex
 * - Full recovery: BackupKeyRecoveryTransansaction with finalized tx hex
 *
 * @param coin - The coin instance
 * @param psbt - The PSBT to format (at appropriate signing stage)
 * @param params - Formatting parameters
 * @returns The formatted recovery result
 */
function formatBackupKeyRecoveryResult(coin, psbt, params) {
    const { walletKeys, keys, recoveryDestination, krsProvider, backupKey, unspents } = params;
    const userHasPrivateKey = hasPrivateKey(keys[0]);
    const backupHasPrivateKey = hasPrivateKey(keys[1]);
    const isUnsignedSweep = !userHasPrivateKey && !backupHasPrivateKey;
    const isKrsRecovery = krsProvider !== undefined && userHasPrivateKey && !backupHasPrivateKey;
    const isFullRecovery = userHasPrivateKey && backupHasPrivateKey;
    // Unsigned sweep - return FormattedOfflineVaultTxInfo
    if (isUnsignedSweep) {
        return {
            txHex: (0, decode_1.encodeTransaction)(psbt).toString('hex'),
            txInfo: {},
            feeInfo: {},
            coin: coin.getChain(),
        };
    }
    const responseTxFormat = !isKrsRecovery || krsProvider === 'keyternal' ? 'legacy' : 'psbt';
    const txInfo = {};
    // Include inputs array for legacy format responses
    txInfo.inputs =
        responseTxFormat === 'legacy'
            ? unspents.map((u) => ({ ...u, value: Number(u.value), valueString: u.value.toString(), prevTx: undefined }))
            : undefined;
    if (isKrsRecovery) {
        // KRS recovery - half-signed
        // keyternal uses legacy format, other KRS providers use PSBT format
        txInfo.transactionHex =
            krsProvider === 'keyternal'
                ? Buffer.from(psbt.getHalfSignedLegacyFormat()).toString('hex')
                : (0, decode_1.encodeTransaction)(psbt).toString('hex');
        txInfo.coin = coin.getChain();
        txInfo.backupKey = backupKey ?? '';
        const recoveryAmount = (0, psbt_1.getRecoveryAmount)(psbt, walletKeys, recoveryDestination);
        txInfo.recoveryAmount = Number(recoveryAmount);
        txInfo.recoveryAmountString = recoveryAmount.toString();
    }
    else if (isFullRecovery) {
        // Full recovery - finalize and extract transaction
        psbt.finalizeAllInputs();
        txInfo.transactionHex = Buffer.from(psbt.extractTransaction().toBytes()).toString('hex');
    }
    return txInfo;
}
function getBip32Keys(bitgo, params) {
    const keys = (0, sdk_core_1.getBip32Keys)(bitgo, params, { requireBitGoXpub: true });
    if (!(0, sdk_core_1.isTriple)(keys)) {
        throw new Error(`expected key triple`);
    }
    return keys.map((k) => wasm_utxo_1.BIP32.from(k.toBase58()));
}
/**
 * Builds a funds recovery transaction without BitGo.
 *
 * Returns transaction hex in legacy format for unsigned sweep transaction, half signed backup recovery transaction with KRS provider (only keyternal),
 * fully signed backup recovery transaction without a KRS provider.
 *
 * Returns PSBT hex for half signed backup recovery transaction with KRS provider (excluding keyternal)
 * For PSBT hex cases, Unspents are not required in response.
 *
 * @param coin
 * @param bitgo
 * @param params
 * - userKey: [encrypted] xprv, or xpub
 * - backupKey: [encrypted] xprv, or xpub if the xprv is held by a KRS provider
 * - walletPassphrase: necessary if one of the xprvs is encrypted
 * - bitgoKey: xpub
 * - krsProvider: necessary if backup key is held by KRS
 * - recoveryDestination: target address to send recovered funds to
 * - scan: the amount of consecutive addresses without unspents to scan through before stopping
 * - ignoreAddressTypes: (optional) scripts to ignore
 *        for example: ['p2shP2wsh', 'p2wsh'] will prevent code from checking for wrapped-segwit and native-segwit chains on the public block explorers
 */
async function backupKeyRecovery(coin, bitgo, params) {
    if (lodash_1.default.isUndefined(params.userKey)) {
        throw new Error('missing userKey');
    }
    if (lodash_1.default.isUndefined(params.backupKey)) {
        throw new Error('missing backupKey');
    }
    if (lodash_1.default.isUndefined(params.recoveryDestination) ||
        !coin.isValidAddress(params.recoveryDestination, { anyFormat: true })) {
        throw new Error('invalid recoveryDestination');
    }
    if (!lodash_1.default.isUndefined(params.scan) && (!lodash_1.default.isInteger(params.scan) || params.scan < 0)) {
        throw new Error('scan must be a positive integer');
    }
    if (params.feeRate !== undefined && (!Number.isFinite(params.feeRate) || params.feeRate <= 0)) {
        throw new Error('feeRate must be a positive number');
    }
    // check whether key material and password authenticate the users and return parent keys of all three keys of the wallet
    const keys = getBip32Keys(bitgo, params);
    const walletKeys = wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from({
        triple: keys,
        derivationPrefixes: [params.userKeyPath || 'm/0/0', 'm/0/0', 'm/0/0'],
    });
    const unspents = (await Promise.all(wasm_utxo_1.fixedScriptWallet.outputScriptTypes
        .filter((addressType) => wasm_utxo_1.fixedScriptWallet.supportsScriptType(coin.name, addressType) &&
        !params.ignoreAddressTypes?.includes(addressType))
        .reduce((queries, addressType) => [
        ...queries,
        queryBlockchainUnspentsPath(coin, params, walletKeys, wasm_utxo_1.fixedScriptWallet.ChainCode.value(addressType, 'external')),
        queryBlockchainUnspentsPath(coin, params, walletKeys, wasm_utxo_1.fixedScriptWallet.ChainCode.value(addressType, 'internal')),
    ], []))).flat();
    const feeRateSatVB = params.feeRate !== undefined
        ? params.feeRate
        : await getRecoveryFeePerBytes(coin, { defaultValue: exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V2 });
    // Calculate KRS fee if needed
    const userHasPrivateKey = hasPrivateKey(keys[0]);
    const backupHasPrivateKey = hasPrivateKey(keys[1]);
    const isKrsRecovery = params.krsProvider !== undefined && userHasPrivateKey && !backupHasPrivateKey;
    let krsFee = BigInt(0);
    let krsFeeAddress;
    if (isKrsRecovery && params.krsProvider) {
        try {
            krsFee = BigInt(await calculateFeeAmount(coin, { provider: params.krsProvider }));
        }
        catch (err) {
            // Don't let this error block the recovery -
            console.dir(err);
        }
        if (krsFee > BigInt(0)) {
            const krsProviderConfig = (0, sdk_core_1.getKrsProvider)(coin, params.krsProvider);
            if (!krsProviderConfig.feeAddresses) {
                throw new Error(`keyProvider must define feeAddresses`);
            }
            krsFeeAddress = krsProviderConfig.feeAddresses[coin.getChain()];
            if (!krsFeeAddress) {
                throw new Error('this KRS provider has not configured their fee structure yet - recovery cannot be completed');
            }
        }
    }
    // Build and sign PSBT
    const psbt = backupKeyRecoveryWithWalletUnspents(coin.name, {
        walletKeys,
        keys,
        recoveryDestination: params.recoveryDestination,
        feeRateSatVB,
        krsFee,
        krsFeeAddress,
    }, unspents);
    // Format the result
    return formatBackupKeyRecoveryResult(coin, psbt, {
        walletKeys,
        keys,
        recoveryDestination: params.recoveryDestination,
        krsProvider: params.krsProvider,
        backupKey: params.backupKey,
        unspents,
    });
}
async function v1BackupKeyRecovery(coin, bitgo, params) {
    if (lodash_1.default.isUndefined(params.recoveryDestination) ||
        !coin.isValidAddress(params.recoveryDestination, { anyFormat: true })) {
        throw new Error('invalid recoveryDestination');
    }
    const recoveryFeePerByte = await getRecoveryFeePerBytes(coin, {
        defaultValue: exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V1,
    });
    const v1wallet = await bitgo.wallets().get({ id: params.walletId });
    return await v1wallet.recover({
        ...params,
        feeRate: recoveryFeePerByte,
    });
}
async function v1Sweep(coin, bitgo, params) {
    if (lodash_1.default.isUndefined(params.recoveryDestination) ||
        !coin.isValidAddress(params.recoveryDestination, { anyFormat: true })) {
        throw new Error('invalid recoveryDestination');
    }
    let recoveryFeePerByte = 100;
    if (bitgo.env === 'prod') {
        recoveryFeePerByte = await getRecoveryFeePerBytes(coin, {
            defaultValue: exports.DEFAULT_RECOVERY_FEERATE_SAT_VBYTE_V1_SWEEP,
        });
    }
    const v1wallet = await bitgo.wallets().get({ id: params.walletId });
    return await v1wallet.sweep({
        ...params,
        feeRate: recoveryFeePerByte,
    });
}
