// @bitgo/abstract-utxo/dist/cjs/src/recovery/baseApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseApi = exports.ApiRequestError = exports.ApiNotImplementedError = void 0;
const superagent_1 = __importDefault(require("superagent"));
const sdk_core_1 = require("@bitgo/sdk-core");
class ApiNotImplementedError extends sdk_core_1.BitGoJsError {
    constructor(coinName) {
        super(`api not implemented for coin ${coinName}`);
    }
}
exports.ApiNotImplementedError = ApiNotImplementedError;
class ApiRequestError extends sdk_core_1.BitGoJsError {
    constructor(url, reason) {
        super(`error in API request: ${url}: ${reason}`);
    }
    static forRequest(req, reason) {
        return new ApiRequestError(req.toString(), reason);
    }
    static forResponse(res, reason) {
        return new ApiRequestError(res.header, reason);
    }
}
exports.ApiRequestError = ApiRequestError;
class BaseApi {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
        if (!baseUrl.startsWith('https://')) {
            throw new Error(`baseUrl must start with https://`);
        }
        if (baseUrl.endsWith(`/`)) {
            throw new Error(`baseUrl must not end with slash (/)`);
        }
    }
    async request(method, path, requestBody, params = {}) {
        if (!path.startsWith(path)) {
            throw new Error(`path must start with '/' character`);
        }
        const url = this.baseUrl + path;
        let response;
        try {
            response = await (0, superagent_1.default)(method, url)
                .retry(params.retry)
                .send(requestBody);
        }
        catch (e) {
            throw new ApiRequestError(url, e);
        }
        if (!response.ok) {
            throw new ApiRequestError(url, `status code ${response.statusCode}`);
        }
        return {
            map(f) {
                try {
                    return f(response.body);
                }
                catch (e) {
                    throw new ApiRequestError(url, e);
                }
            },
        };
    }
    get(path, params) {
        return this.request('get', path, undefined, params);
    }
    post(path, body, params) {
        return this.request('post', path, body, params);
    }
}
exports.BaseApi = BaseApi;
