// @bitgo/abstract-utxo/dist/cjs/src/recovery/psbt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTaprootChain = isTaprootChain;
exports.createEmptyWasmPsbt = createEmptyWasmPsbt;
exports.addWalletInputsToWasmPsbt = addWalletInputsToWasmPsbt;
exports.addOutputToWasmPsbt = addOutputToWasmPsbt;
exports.createBackupKeyRecoveryPsbt = createBackupKeyRecoveryPsbt;
exports.getRecoveryAmount = getRecoveryAmount;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const unspent_1 = require("../unspent");
/**
 * Check if a chain code is for a taproot script type
 */
function isTaprootChain(chain) {
    const scriptType = wasm_utxo_1.fixedScriptWallet.ChainCode.scriptType(chain);
    return scriptType === 'p2trLegacy' || scriptType === 'p2trMusig2';
}
class InsufficientFundsError extends Error {
    constructor(totalInputAmount, approximateFee, krsFee, recoveryAmount) {
        super(`This wallet's balance is too low to pay the fees specified by the KRS provider.` +
            `Existing balance on wallet: ${totalInputAmount.toString()}. ` +
            `Estimated network fee for the recovery transaction: ${approximateFee.toString()}` +
            `KRS fee to pay: ${krsFee.toString()}. ` +
            `After deducting fees, your total recoverable balance is ${recoveryAmount.toString()}`);
        this.totalInputAmount = totalInputAmount;
        this.approximateFee = approximateFee;
        this.krsFee = krsFee;
        this.recoveryAmount = recoveryAmount;
    }
}
/**
 * Check if the network is a Zcash network
 */
function isZcash(coinName) {
    return coinName === 'zec' || coinName === 'tzec';
}
/**
 * Default block heights for Zcash networks if not provided.
 * These should be set to a height after the latest network upgrade.
 * TODO(BTC-2901): get the height from blockchair API instead of hardcoding.
 */
const ZCASH_DEFAULT_BLOCK_HEIGHTS = {
    zec: 3146400,
    tzec: 3536500,
};
/**
 * Create an empty wasm-utxo BitGoPsbt for a given network.
 * Handles Zcash networks specially by using ZcashBitGoPsbt.
 *
 * @param network - The network for the PSBT
 * @param rootWalletKeys - The wallet keys
 * @param options - Optional settings (e.g., blockHeight for Zcash)
 * @returns A wasm-utxo BitGoPsbt instance
 */
function createEmptyWasmPsbt(coinName, rootWalletKeys, options) {
    if (isZcash(coinName)) {
        // For Zcash, use ZcashBitGoPsbt which requires block height to determine consensus branch ID
        const blockHeight = options?.blockHeight ?? ZCASH_DEFAULT_BLOCK_HEIGHTS[coinName];
        return wasm_utxo_1.fixedScriptWallet.ZcashBitGoPsbt.createEmpty(coinName, rootWalletKeys, {
            blockHeight,
        });
    }
    return wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty(coinName, rootWalletKeys);
}
/**
 * Add wallet inputs from unspents to a wasm-utxo BitGoPsbt.
 * Handles taproot inputs by setting the appropriate signPath.
 *
 * @param wasmPsbt - The wasm-utxo BitGoPsbt to add inputs to
 * @param unspents - The wallet unspents to add as inputs
 * @param rootWalletKeys - The wallet keys
 */
function addWalletInputsToWasmPsbt(wasmPsbt, unspents, rootWalletKeys) {
    unspents.forEach((unspent) => {
        const { txid, vout } = (0, unspent_1.parseOutputId)(unspent.id);
        const signPath = isTaprootChain(unspent.chain)
            ? { signer: 'user', cosigner: 'backup' }
            : undefined;
        // prevTx may be added dynamically in backupKeyRecovery for non-segwit inputs
        const prevTx = unspent.prevTx;
        wasmPsbt.addWalletInput({
            txid,
            vout,
            value: unspent.value,
            prevTx: prevTx,
        }, rootWalletKeys, {
            scriptId: { chain: unspent.chain, index: unspent.index },
            signPath,
        });
    });
}
/**
 * Add an output to a wasm-utxo BitGoPsbt.
 *
 * @param wasmPsbt - The wasm-utxo BitGoPsbt to add the output to
 * @param address - The destination address
 * @param value - The output value in satoshis
 * @param network - The network (used to convert address to script)
 * @returns The output index
 */
function addOutputToWasmPsbt(wasmPsbt, address, value, coinName) {
    const script = wasm_utxo_1.address.toOutputScriptWithCoin(address, coinName);
    return wasmPsbt.addOutput({ script, value });
}
/**
 * Create a backup key recovery PSBT using wasm-utxo
 */
function createBackupKeyRecoveryPsbtWasm(coinName, rootWalletKeys, unspents, options) {
    const { feeRateSatVB, recoveryDestination, keyRecoveryServiceFee, keyRecoveryServiceFeeAddress } = options;
    // Create PSBT with wasm-utxo and add wallet inputs using shared utilities
    const wasmPsbt = createEmptyWasmPsbt(coinName, rootWalletKeys, { blockHeight: options.blockHeight });
    addWalletInputsToWasmPsbt(wasmPsbt, unspents, rootWalletKeys);
    // Calculate dimensions using wasm-utxo Dimensions
    let dimensions = wasm_utxo_1.fixedScriptWallet.Dimensions.fromPsbt(wasmPsbt).plus(wasm_utxo_1.fixedScriptWallet.Dimensions.fromOutput(recoveryDestination, coinName));
    if (keyRecoveryServiceFeeAddress) {
        dimensions = dimensions.plus(wasm_utxo_1.fixedScriptWallet.Dimensions.fromOutput(keyRecoveryServiceFeeAddress, coinName));
    }
    const approximateFee = BigInt(dimensions.getVSize() * feeRateSatVB);
    const totalInputAmount = (0, unspent_1.unspentSum)(unspents);
    const recoveryAmount = totalInputAmount - approximateFee - keyRecoveryServiceFee;
    if (recoveryAmount < BigInt(0)) {
        throw new InsufficientFundsError(totalInputAmount, approximateFee, keyRecoveryServiceFee, recoveryAmount);
    }
    // Add outputs to wasm PSBT
    addOutputToWasmPsbt(wasmPsbt, recoveryDestination, recoveryAmount, coinName);
    if (keyRecoveryServiceFeeAddress) {
        addOutputToWasmPsbt(wasmPsbt, keyRecoveryServiceFeeAddress, keyRecoveryServiceFee, coinName);
    }
    return wasmPsbt;
}
/**
 * Create a backup key recovery PSBT.
 *
 * @param coinName - The coin name for the PSBT
 * @param rootWalletKeys - The wallet keys
 * @param unspents - The unspents to include in the PSBT
 * @param options - Options for creating the PSBT
 */
function createBackupKeyRecoveryPsbt(coinName, rootWalletKeys, unspents, options) {
    if (options.keyRecoveryServiceFee > 0 && !options.keyRecoveryServiceFeeAddress) {
        throw new Error('keyRecoveryServiceFeeAddress is required when keyRecoveryServiceFee is provided');
    }
    return createBackupKeyRecoveryPsbtWasm(coinName, rootWalletKeys, unspents, options);
}
function getRecoveryAmount(psbt, walletKeys, address) {
    const parsedOutputs = psbt.parseOutputsWithWalletKeys(walletKeys);
    const recoveryOutput = parsedOutputs.find((o) => o.address === address);
    if (!recoveryOutput) {
        throw new Error(`Recovery destination output not found in PSBT: ${address}`);
    }
    return BigInt(recoveryOutput.value);
}
