// @bitgo/abstract-utxo/dist/cjs/test/unit/recovery/backupKeyRecoveryUnspentGathering.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const sinon = __importStar(require("sinon"));
const nock = require("nock");
const src_1 = require("../../../src");
const util_1 = require("../util");
const mock_1 = require("./mock");
nock.disableNetConnect();
const walletPassphrase = 'lol';
// Get default wasm wallet keys (xpubs, xprivs, and walletKeys)
const { walletKeys: wasmWalletKeys, xpubs, xprivs } = (0, util_1.getDefaultWasmWalletKeys)();
function getNamedKeys(keys, password) {
    function encode(k) {
        const base58 = k.toBase58();
        // Check if it's a public key
        const pubKeyMatch = util_1.keychainsBase58.find((kc) => kc.pub === base58);
        if (pubKeyMatch) {
            return base58;
        }
        // It's a private key - find and encrypt it
        const keyBase58 = util_1.keychainsBase58.find((kc) => kc.prv === base58);
        if (!keyBase58) {
            throw new Error('Key not found in keychainsBase58');
        }
        return (0, util_1.encryptKeychain)(password, keyBase58);
    }
    return {
        userKey: encode(keys[0]),
        backupKey: encode(keys[1]),
        bitgoKey: encode(keys[2]),
    };
}
function getKeysForFullSignedRecovery(password) {
    return getNamedKeys([xprivs[0], xprivs[1], xpubs[2]], password);
}
const keysFullSignedRecovery = getKeysForFullSignedRecovery(walletPassphrase);
/**
 * Tests for unspent gathering via backupKeyRecovery with MockRecoveryProvider.
 * This validates the address scanning and unspent collection logic separately
 * from the transaction building logic tested in backupKeyRecovery.ts.
 */
describe('Backup Key Recovery - Unspent Gathering', function () {
    const defaultFeeRateSatB = 100;
    (0, util_1.getMinUtxoCoins)().forEach((coin) => {
        describe(`Unspent Gathering [${coin.getChain()}]`, function () {
            const { walletKeys: externalWallet } = (0, util_1.createWasmWalletKeys)('external');
            const recoveryDestination = (0, util_1.getWalletAddress)(coin.name, externalWallet);
            before('mock', function () {
                sinon.stub(src_1.CoingeckoApi.prototype, 'getUSDPrice').resolves(69420);
            });
            after(function () {
                sinon.restore();
            });
            it('gathers unspents from recovery provider and builds transaction', async function () {
                this.timeout(10000);
                // Use p2sh for all coins (universal support)
                const scriptTypes = ['p2sh'];
                // Create test unspents with prevTx using wasm-utxo only (no utxolib dependency)
                const testUnspents = scriptTypes.flatMap((scriptType, index) => [
                    (0, util_1.toUnspentWithPrevTx)({ scriptType, value: BigInt(1e8) }, index, coin.name, wasmWalletKeys),
                    (0, util_1.toUnspentWithPrevTx)({ scriptType, value: BigInt(2e8) }, index, coin.name, wasmWalletKeys),
                ]);
                // Convert to API format for BCH/BCHA if needed
                const mockedApiUnspents = coin.getChain() === 'bch' || coin.getChain() === 'bcha'
                    ? testUnspents.map((u) => ({
                        ...u,
                        address: coin.canonicalAddress(u.address, 'cashaddr').split(':')[1],
                    }))
                    : testUnspents;
                const mockProvider = new mock_1.MockRecoveryProvider(mockedApiUnspents);
                // Verify mock provider works correctly
                const fetchedUnspents = await mockProvider.getUnspentsForAddresses(mockedApiUnspents.map((u) => u.address));
                assert_1.default.strictEqual(fetchedUnspents.length, mockedApiUnspents.length);
                // Call backupKeyRecovery which should gather unspents via the mock provider
                const recovery = await (0, src_1.backupKeyRecovery)(coin, util_1.defaultBitGo, {
                    walletPassphrase,
                    recoveryDestination,
                    scan: 5,
                    ignoreAddressTypes: [],
                    ...keysFullSignedRecovery,
                    recoveryProvider: mockProvider,
                    feeRate: defaultFeeRateSatB,
                });
                // Verify recovery succeeded (transaction hex indicates unspents were gathered)
                const txHex = recovery.transactionHex ??
                    recovery.txHex;
                assert_1.default.ok(txHex, 'should have transaction hex from gathered unspents');
            });
            it('handles empty unspents from recovery provider', async function () {
                this.timeout(10000);
                const mockProvider = new mock_1.MockRecoveryProvider([]);
                await assert_1.default.rejects(async () => {
                    await (0, src_1.backupKeyRecovery)(coin, util_1.defaultBitGo, {
                        walletPassphrase,
                        recoveryDestination,
                        scan: 2,
                        ignoreAddressTypes: [],
                        ...keysFullSignedRecovery,
                        recoveryProvider: mockProvider,
                        feeRate: defaultFeeRateSatB,
                    });
                }, /No input to recover/);
            });
        });
    });
});
