// @bitgo/abstract-utxo/dist/cjs/test/unit/recovery/backupKeyRecovery.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
require("should");
const nock = require("nock");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const src_1 = require("../../../src");
const util_1 = require("../util");
nock.disableNetConnect();
const defaultDerivationPrefix = 'm/0/0';
const exoticUserKeyPath = '99/99';
// Create wallet keys using wasm-utxo primitives from default keychains
const { walletKeys: wasmWalletKeys, xprivs } = (0, util_1.getDefaultWasmWalletKeys)();
const { walletKeys: wasmExoticWalletKeys } = (0, util_1.getDefaultWasmWalletKeys)([
    exoticUserKeyPath,
    defaultDerivationPrefix,
    defaultDerivationPrefix,
]);
// Private keys for signing
const userPrivkey = xprivs[0];
const backupPrivkey = xprivs[1];
function run(coin, scriptTypes, wasmWalletKeys, params, tags = []) {
    const defaultFeeRateSatB = 100;
    describe(`Backup Key Recovery PSBT [${[coin.getChain(), ...tags].join(',')}]`, function () {
        const { walletKeys: externalWallet } = (0, util_1.createWasmWalletKeys)('external');
        const recoveryDestination = (0, util_1.getWalletAddress)(coin.name, externalWallet);
        const fixtureCoin = (0, util_1.getNormalTestnetCoin)(coin);
        // Get xpubs from wallet keys
        const userKey = wasmWalletKeys.userKey();
        const backupKey = wasmWalletKeys.backupKey();
        const bitgoKey = wasmWalletKeys.bitgoKey();
        const replayProtection = [wasm_utxo_1.ECPair.fromPublicKey(Buffer.from(userKey.publicKey))];
        // 1e8 * 9e7 < 9.007e15 but 2e8 * 9e7 > 9.007e15 to test both code paths in queryBlockchainUnspentsPath
        const valueMul = coin.amountType === 'bigint' ? BigInt(9e7) : BigInt(1);
        let psbt;
        let recoverUnspents;
        let parsed;
        let fixtureParsed;
        before('create recovery data and load fixture', async function () {
            this.timeout(10000);
            recoverUnspents = scriptTypes.flatMap((scriptType, index) => [
                (0, util_1.toUnspent)({ scriptType, value: BigInt(1e8) * valueMul }, index, coin.name, wasmWalletKeys),
                (0, util_1.toUnspent)({ scriptType, value: BigInt(2e8) * valueMul }, index, coin.name, wasmWalletKeys),
                (0, util_1.toUnspent)({ scriptType, value: BigInt(3e8) * valueMul }, index, coin.name, wasmWalletKeys),
            ]);
            // Build signing keys triple: privkey if provided, otherwise xpub from wallet keys
            const signingKeys = [
                params.userPrivkey || userKey,
                params.backupPrivkey || backupKey,
                bitgoKey, // bitgo is always xpub
            ];
            psbt = (0, src_1.backupKeyRecoveryWithWalletUnspents)(coin.name, {
                walletKeys: wasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                feeRateSatVB: defaultFeeRateSatB,
            }, recoverUnspents);
            // Parse generated PSBT
            parsed = psbt.parseTransactionWithWalletKeys(wasmWalletKeys, {
                replayProtection: { publicKeys: replayProtection },
            });
            // Load and parse fixture PSBT
            const psbtHex = Buffer.from(psbt.serialize()).toString('hex');
            const storedFixture = await (0, util_1.getFixture)(fixtureCoin, `recovery/backupKeyRecoveryPsbt-${tags.join('-')}`, {
                psbtHex,
            });
            const fixturePsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(Buffer.from(storedFixture.psbtHex, 'hex'), fixtureCoin.name);
            fixtureParsed = fixturePsbt.parseTransactionWithWalletKeys(wasmWalletKeys, {
                replayProtection: { publicKeys: replayProtection },
            });
        });
        it('has expected input count', function () {
            parsed.inputs.length.should.eql(recoverUnspents.length);
        });
        it('has recovery destination output', function () {
            parsed.outputs.length.should.be.greaterThanOrEqual(1);
            const outputAddresses = parsed.outputs.map((o) => o.address);
            outputAddresses.includes(recoveryDestination).should.eql(true);
        });
        it('has expected fee rate', function () {
            const inputSum = recoverUnspents.reduce((sum, u) => sum + u.value, BigInt(0));
            const outputSum = parsed.outputs.reduce((sum, o) => sum + o.value, BigInt(0));
            const fee = inputSum - outputSum;
            const vsize = wasm_utxo_1.fixedScriptWallet.Dimensions.fromPsbt(psbt).getVSize();
            const feeRateSatB = Number(fee) / vsize;
            const diff = Math.abs(feeRateSatB - defaultFeeRateSatB) / defaultFeeRateSatB;
            // within 1%
            assert_1.default.strictEqual(diff < 0.01, true, `expected fee rate ${defaultFeeRateSatB} but got ${feeRateSatB}`);
        });
        function checkInputsSignedBy(signerIndex, expectSigned) {
            const signerKey = xprivs[signerIndex];
            for (let inputIndex = 0; inputIndex < recoverUnspents.length; inputIndex++) {
                const hasSig = psbt.verifySignature(inputIndex, signerKey);
                hasSig.should.eql(expectSigned, `input ${inputIndex} signer ${signerIndex}`);
            }
        }
        it((params.userPrivkey ? 'has' : 'has no') + ' user signature', function () {
            checkInputsSignedBy(0, !!params.userPrivkey);
        });
        it((params.backupPrivkey ? 'has' : 'has no') + ' backup signature', function () {
            checkInputsSignedBy(1, !!params.backupPrivkey);
        });
        it('matches PSBT fixture', function () {
            assert_1.default.deepStrictEqual(parsed, fixtureParsed);
        });
    });
}
function runWithScriptTypes(scriptTypes) {
    util_1.utxoCoins
        .filter((coin) => scriptTypes.every((type) => coin.supportsAddressType(type)))
        .forEach((coin) => {
        // Unsigned sweep - no signatures
        run(coin, scriptTypes, wasmWalletKeys, {}, ['unsigned', ...scriptTypes]);
        // User signed only (KRS scenario)
        run(coin, scriptTypes, wasmWalletKeys, { userPrivkey }, ['userSigned', ...scriptTypes]);
        // Fully signed recovery
        run(coin, scriptTypes, wasmWalletKeys, { userPrivkey, backupPrivkey }, ['fullySigned', ...scriptTypes]);
        // Fully signed with custom user key path
        run(coin, scriptTypes, wasmExoticWalletKeys, { userPrivkey, backupPrivkey }, [
            'fullySigned',
            'customUserKeyPath',
            ...scriptTypes,
        ]);
    });
}
describe('Backup Key Recovery PSBT', function () {
    // compatible with all coins
    runWithScriptTypes(['p2sh']);
    // segwit compatible coins
    runWithScriptTypes(['p2shP2wsh', 'p2wsh']);
    // taproot compatible coins
    runWithScriptTypes(['p2tr', 'p2trMusig2']);
});
