// @bitgo/abstract-utxo/dist/cjs/test/unit/recovery/crossChainRecovery.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const should = require("should");
const nock = require("nock");
const sdk_test_1 = require("@bitgo/sdk-test");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const sinon = __importStar(require("sinon"));
const src_1 = require("../../../src");
const names_1 = require("../../../src/names");
const util_1 = require("../util");
const nockBitGo_1 = require("../util/nockBitGo");
const mock_1 = require("./mock");
function getKeyId(k) {
    return (0, sdk_test_1.getSeed)(k.pub).toString('hex');
}
function nockWallet(coin, walletId, walletKeys) {
    return [
        (0, nockBitGo_1.nockBitGo)()
            .get(`/api/v2/${coin.getChain()}/wallet/${walletId}`)
            .reply(200, {
            id: walletId,
            coin: coin.getChain(),
            label: 'crossChainRecovery',
            keys: walletKeys.map((k) => getKeyId(k)),
        })
            .persist(),
        ...walletKeys.map((k) => (0, nockBitGo_1.nockBitGo)()
            .get(`/api/v2/${coin.getChain()}/key/${getKeyId(k)}`)
            .reply(200, k)
            .persist()),
    ];
}
function nockWalletAddress(coin, walletId, address) {
    return (0, nockBitGo_1.nockBitGo)()
        .get(`/api/v2/${coin.getChain()}/wallet/${walletId}/address/${address.address}`)
        .reply(200, {
        address: address.address,
        chain: address.chain,
        index: address.index,
        coin: coin.getChain(),
        wallet: walletId,
    })
        .persist();
}
/**
 * Setup test for cross-chain recovery.
 *
 * Users can receive deposits on wallet addresses that are on a different chain.
 *
 * For instance, a user can receive litecoin on a bitcoin wallet.
 * This means that the litecoin blockchain has a transaction with outputs that are spendable
 * with keys that were originally created for a BitGo BTC wallet.
 * In this example, LTC is the "source coin" and BTC is the "recovery coin"
 * In cases like these we must use construct a transaction for litecoin network using keys of the
 * bitcoin wallet.
 *
 * @param sourceCoin - the coin to construct the transaction for
 * @param recoveryCoin - the coin the receiving wallet was set up for
 */
// Get wasm wallet keys for signature verification
const { walletKeys: wasmWalletKeys, xprivs } = (0, util_1.getDefaultWasmWalletKeys)();
function run(sourceCoin, recoveryCoin) {
    describe(`Cross-Chain Recovery [sourceCoin=${sourceCoin.getChain()} recoveryCoin=${recoveryCoin.getChain()}]`, function () {
        const recoveryWalletId = '5abacebe28d72fbd07e0b8cbba0ff39e';
        // the address the accidental deposit went to, in both sourceCoin and addressCoin formats
        const [depositAddressSourceCoin, depositAddressRecoveryCoin] = [sourceCoin, recoveryCoin].map((coin) => ({
            address: (0, src_1.generateAddress)(coin.name, { keychains: util_1.keychainsBase58, chain: 0, index: 0 }),
            chain: 0,
            index: 0,
        }));
        const chain = 0;
        const index = 1;
        // the address where we want to recover our funds to
        const recoveryAddress = (0, src_1.generateAddress)(sourceCoin.name, {
            keychains: util_1.keychainsBase58,
            chain,
            index,
        });
        const nocks = [];
        // Create recovery unspent using wasm-utxo (no utxolib dependency)
        function getRecoveryUnspents() {
            const depositValue = (sourceCoin.amountType === 'bigint' ? BigInt(1e8) : 1e8);
            return [
                (0, util_1.createWasmWalletUnspent)(depositAddressSourceCoin.address, chain, index, depositValue, sourceCoin.name),
            ];
        }
        before('setup nocks', function () {
            nocks.push(...nockWallet(recoveryCoin, recoveryWalletId, util_1.keychainsBase58));
            nocks.push(nockWalletAddress(recoveryCoin, recoveryWalletId, depositAddressRecoveryCoin));
        });
        after(function () {
            nocks.forEach((n) => n.done());
        });
        after(function () {
            nock.cleanAll();
        });
        afterEach(function () {
            sinon.restore();
        });
        async function matchFixture(name, recovery) {
            let recoveryObj = { ...recovery };
            if (sourceCoin.amountType === 'bigint') {
                recoveryObj = JSON.parse(JSON.stringify(recoveryObj, (k, v) => {
                    if (typeof v === 'bigint') {
                        return v.toString();
                    }
                    else {
                        return v;
                    }
                }));
            }
            (0, util_1.shouldEqualJSON)(recoveryObj, await (0, util_1.getFixture)(sourceCoin, `recovery/crossChainRecovery-${recoveryCoin.getChain()}-${name}`, recoveryObj));
        }
        function checkRecoveryPsbtSignature(psbtHex) {
            // Parse using wasm-utxo for signature verification
            const wasmPsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(Buffer.from(psbtHex, 'hex'), sourceCoin.name);
            const parsed = wasmPsbt.parseTransactionWithWalletKeys(wasmWalletKeys, { replayProtection: { publicKeys: [] } });
            const unspents = getRecoveryUnspents();
            should.equal(parsed.inputs.length, unspents.length);
            // Verify user key has signed each input (same pattern as backupKeyRecovery test)
            parsed.inputs.forEach((_, i) => {
                const userSigned = wasmPsbt.verifySignature(i, xprivs[0]);
                userSigned.should.eql(true, `Input ${i} should be signed by user key`);
            });
        }
        it('should test signed cross chain recovery', async () => {
            const depositUnspent = getRecoveryUnspents()[0];
            // Parse txid from the unspent id (format: txid:vout)
            const depositTxid = depositUnspent.id.split(':')[0];
            // Use the deposit address as the input address (simulating a p2sh input)
            const inputAddress = (0, src_1.generateAddress)(sourceCoin.name, { keychains: util_1.keychainsBase58, chain: 0, index: 2 });
            const getRecoveryProviderStub = sinon
                .stub(src_1.AbstractUtxoCoin.prototype, 'getRecoveryProvider')
                .returns(new mock_1.WasmCrossChainRecoveryProvider(sourceCoin, depositUnspent, [inputAddress]));
            const params = {
                recoveryCoin,
                txid: depositTxid,
                recoveryAddress,
                wallet: recoveryWalletId,
            };
            const signedRecovery = (await sourceCoin.recoverFromWrongChain({
                ...params,
                xprv: util_1.keychainsBase58[0].prv,
            }));
            should.equal(getRecoveryProviderStub.callCount, 1);
            // Verify fixture match
            await matchFixture('signed', signedRecovery);
            // Verify PSBT has valid signatures (user key signed)
            checkRecoveryPsbtSignature(signedRecovery.txHex);
        });
        it('should test unsigned cross chain recovery', async () => {
            const depositUnspent = getRecoveryUnspents()[0];
            // Parse txid from the unspent id (format: txid:vout)
            const depositTxid = depositUnspent.id.split(':')[0];
            // Use the deposit address as the input address (simulating a p2sh input)
            const inputAddress = (0, src_1.generateAddress)(sourceCoin.name, { keychains: util_1.keychainsBase58, chain: 0, index: 2 });
            const getRecoveryProviderStub = sinon
                .stub(src_1.AbstractUtxoCoin.prototype, 'getRecoveryProvider')
                .returns(new mock_1.WasmCrossChainRecoveryProvider(sourceCoin, depositUnspent, [inputAddress]));
            const params = {
                recoveryCoin,
                txid: depositTxid,
                recoveryAddress,
                wallet: recoveryWalletId,
            };
            const unsignedRecovery = (await sourceCoin.recoverFromWrongChain({
                ...params,
                signed: false,
            }));
            should.equal(getRecoveryProviderStub.callCount, 1);
            // Verify fixture match
            await matchFixture('unsigned', unsignedRecovery);
            // Verify the unsigned PSBT can be signed
            const signedTx = await sourceCoin.signTransaction({
                txPrebuild: unsignedRecovery,
                prv: util_1.keychainsBase58[0].prv,
                pubs: util_1.keychainsBase58.map((k) => k.pub),
            });
            checkRecoveryPsbtSignature(signedTx.txHex);
        });
    });
}
function isSupportedCrossChainRecovery(sourceCoin, recoveryCoin) {
    return src_1.supportedCrossChainRecoveries[sourceCoin.getFamily()]?.includes(recoveryCoin.getFamily());
}
util_1.utxoCoins.forEach((coin) => {
    util_1.utxoCoins
        .filter((otherCoin) => coin !== otherCoin &&
        isSupportedCrossChainRecovery(coin, otherCoin) &&
        (((0, names_1.isMainnetCoin)(coin.name) && (0, names_1.isMainnetCoin)(otherCoin.name)) ||
            ((0, names_1.isTestnetCoin)(coin.name) && (0, names_1.isTestnetCoin)(otherCoin.name))))
        .forEach((otherCoin) => {
        if (coin.amountType === 'bigint') {
            run(coin, otherCoin);
        }
        else {
            run(coin, otherCoin);
        }
    });
});
describe(`Cross-Chain Recovery getWallet`, async function () {
    const bitgo = util_1.defaultBitGo;
    const recoveryCoin = (0, util_1.getUtxoCoin)('btc');
    const recoveryWalletId = '5abacebe28d72fbd07e0b8cbba0ff39e';
    it('should search v1 wallets if the v2 endpoint responds with a 4xx error', async function () {
        const errorResponses = [400, 404];
        for (const error of errorResponses) {
            const nockV2Wallet = (0, nockBitGo_1.nockBitGo)(bitgo)
                .get(`/api/v2/${recoveryCoin.getChain()}/wallet/${recoveryWalletId}`)
                .reply(error);
            const nockV1Wallet = (0, nockBitGo_1.nockBitGo)(bitgo).get(`/api/v1/wallet/${recoveryWalletId}`).reply(error);
            await assert.rejects(() => (0, src_1.getWallet)(bitgo, recoveryCoin, recoveryWalletId), Error(`could not get wallet ${recoveryWalletId} from v1 or v2: ApiResponseError: ${error}`));
            nockV2Wallet.done();
            nockV1Wallet.done();
        }
    });
    it('should throw an error if the v2 endpoint responds with a 5xx error', async function () {
        const errorResponses = [500];
        for (const error of errorResponses) {
            const nockV2Wallet = (0, nockBitGo_1.nockBitGo)(bitgo)
                .get(`/api/v2/${recoveryCoin.getChain()}/wallet/${recoveryWalletId}`)
                .reply(error);
            await assert.rejects(() => (0, src_1.getWallet)(bitgo, recoveryCoin, recoveryWalletId), {
                name: 'ApiResponseError',
                status: 500,
                result: {},
                invalidToken: false,
                needsOTP: false,
            });
            nockV2Wallet.done();
        }
    });
});
describe('convertLtcAddressToLegacyFormat', function () {
    const scriptPubKey = Buffer.from('a9149f0bf51fab4d33ab21977e1b89f776f64161ef4287', 'hex');
    it('should convert M... P2SH address to 3... legacy format', function () {
        // These two addresses represent the same underlying script hash:
        // - MNQ7zkgMsaV67rsjA3JuP59RC5wxRXpwgE is the LTC format (scriptHash 0x32)
        // - 3GBygsGPvTdfKMbq4AKZZRu1sPMWPEsBfd is the BTC format (scriptHash 0x05)
        // const ltcAddress = 'MNQ7zkgMsaV67rsjA3JuP59RC5wxRXpwgE';
        const ltcAddress = wasm_utxo_1.address.fromOutputScriptWithCoin(scriptPubKey, 'ltc');
        const tltcAddress = wasm_utxo_1.address.fromOutputScriptWithCoin(scriptPubKey, 'tltc');
        assert.strictEqual(ltcAddress, 'MNQ7zkgMsaV67rsjA3JuP59RC5wxRXpwgE');
        assert.strictEqual(tltcAddress, 'Qb6wsd4fZ2C6fKzRMPyTG5KiE81W6QCxhU');
        const legacyAddress = (0, src_1.convertLtcAddressToLegacyFormat)(ltcAddress, 'ltc');
        const expectedLtcLegacyAddress = '3GBygsGPvTdfKMbq4AKZZRu1sPMWPEsBfd';
        assert.strictEqual(legacyAddress, expectedLtcLegacyAddress);
        const expectedTltcLegacyAddress = '2N7kBkcCRXv91X9ENjHwSBNtH5jZg8yye6r';
        const tltcLegacyAddress = (0, src_1.convertLtcAddressToLegacyFormat)(tltcAddress, 'tltc');
        assert.strictEqual(tltcLegacyAddress, expectedTltcLegacyAddress);
    });
    it('should convert MD68PsdheKxcYsrVLyZRXgoSDLnB1MdVtE to legacy format', function () {
        const address = 'MD68PsdheKxcYsrVLyZRXgoSDLnB1MdVtE';
        const legacyAddress = (0, src_1.convertLtcAddressToLegacyFormat)(address, 'ltc');
        // Should start with '3' (legacy BTC P2SH format)
        assert.ok(legacyAddress.startsWith('3'), `Expected address to start with '3', got: ${legacyAddress}`);
    });
    it('should not modify bech32 addresses', function () {
        const bech32Address = 'ltc1qgrl8zpndsklaa9swgd5vevyxmx5x63vcrl7dk4';
        const result = (0, src_1.convertLtcAddressToLegacyFormat)(bech32Address, 'ltc');
        assert.strictEqual(result, bech32Address);
    });
});
