// @bitgo/abstract-utxo/dist/cjs/test/unit/recovery/formatBackupKeyRecoveryResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("should");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const src_1 = require("../../../src");
const util_1 = require("../util");
// Use tbtc for all tests - formatting logic doesn't vary by coin
const coin = util_1.utxoCoins.find((c) => c.getChain() === 'tbtc');
const scriptType = 'p2shP2wsh';
// Get default wasm wallet keys (xpubs and xprivs)
const { walletKeys: defaultWasmWalletKeys, xpubs, xprivs } = (0, util_1.getDefaultWasmWalletKeys)();
/**
 * Build signing keys triple with specified neuter options.
 */
function buildSigningKeys(options) {
    return [
        options.neuterUser ? xpubs[0] : xprivs[0],
        options.neuterBackup ? xpubs[1] : xprivs[1],
        xpubs[2], // bitgo is always xpub
    ];
}
function createTestUnspents() {
    return [
        (0, util_1.toUnspent)({ scriptType, value: BigInt(1e8) }, 0, coin.name, defaultWasmWalletKeys),
        (0, util_1.toUnspent)({ scriptType, value: BigInt(2e8) }, 0, coin.name, defaultWasmWalletKeys),
    ];
}
/**
 * Clone a PSBT - necessary because formatBackupKeyRecoveryResult mutates when finalizing.
 */
function clonePsbt(psbt) {
    return wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(psbt.serialize(), coin.name);
}
describe('formatBackupKeyRecoveryResult', function () {
    const { walletKeys: externalWallet } = (0, util_1.createWasmWalletKeys)('external');
    const recoveryDestination = (0, util_1.getWalletAddress)(coin.name, externalWallet);
    const feeRateSatVB = 100;
    describe('unsigned sweep (no signatures)', function () {
        let psbt;
        let signingKeys;
        let unspents;
        before(function () {
            unspents = createTestUnspents();
            signingKeys = buildSigningKeys({ neuterUser: true, neuterBackup: true });
            psbt = (0, src_1.backupKeyRecoveryWithWalletUnspents)(coin.name, { walletKeys: defaultWasmWalletKeys, keys: signingKeys, recoveryDestination, feeRateSatVB }, unspents);
        });
        it('returns FormattedOfflineVaultTxInfo with PSBT hex', function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                unspents,
            });
            result.should.have.property('txHex');
            result.should.have.property('txInfo');
            result.should.have.property('feeInfo');
            result.should.have.property('coin');
            result.coin.should.equal(coin.getChain());
            const txBuf = Buffer.from(result.txHex, 'hex');
            (0, wasm_utxo_1.hasPsbtMagic)(txBuf).should.equal(true);
        });
        it('matches fixture', async function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                unspents,
            });
            const fixture = await (0, util_1.getFixture)(coin, 'recovery/formatBackupKeyRecovery-unsigned', result);
            (0, util_1.shouldEqualJSON)(result, fixture);
        });
    });
    describe('KRS recovery with keyternal (legacy format)', function () {
        let psbt;
        let signingKeys;
        let unspents;
        before(function () {
            unspents = createTestUnspents();
            signingKeys = buildSigningKeys({ neuterUser: false, neuterBackup: true });
            psbt = (0, src_1.backupKeyRecoveryWithWalletUnspents)(coin.name, { walletKeys: defaultWasmWalletKeys, keys: signingKeys, recoveryDestination, feeRateSatVB }, unspents);
        });
        it('returns BackupKeyRecoveryTransansaction with legacy tx hex', function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                krsProvider: 'keyternal',
                backupKey: 'test-backup-key',
                unspents,
            });
            result.should.have.property('transactionHex');
            result.should.have.property('coin');
            result.should.have.property('backupKey');
            result.should.have.property('recoveryAmount');
            result.should.have.property('inputs');
            result.backupKey.should.equal('test-backup-key');
            const txBuf = Buffer.from(result.transactionHex, 'hex');
            (0, wasm_utxo_1.hasPsbtMagic)(txBuf).should.equal(false);
        });
        it('matches fixture', async function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                krsProvider: 'keyternal',
                backupKey: 'test-backup-key',
                unspents,
            });
            const fixture = await (0, util_1.getFixture)(coin, 'recovery/formatBackupKeyRecovery-krs-keyternal', result);
            (0, util_1.shouldEqualJSON)(result, fixture);
        });
    });
    describe('KRS recovery with dai (PSBT format)', function () {
        let psbt;
        let signingKeys;
        let unspents;
        before(function () {
            unspents = createTestUnspents();
            signingKeys = buildSigningKeys({ neuterUser: false, neuterBackup: true });
            psbt = (0, src_1.backupKeyRecoveryWithWalletUnspents)(coin.name, { walletKeys: defaultWasmWalletKeys, keys: signingKeys, recoveryDestination, feeRateSatVB }, unspents);
        });
        it('returns BackupKeyRecoveryTransansaction with PSBT hex', function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                krsProvider: 'dai',
                backupKey: 'test-backup-key',
                unspents,
            });
            result.should.have.property('transactionHex');
            result.should.have.property('coin');
            result.should.have.property('backupKey');
            result.should.have.property('recoveryAmount');
            result.backupKey.should.equal('test-backup-key');
            const txBuf = Buffer.from(result.transactionHex, 'hex');
            (0, wasm_utxo_1.hasPsbtMagic)(txBuf).should.equal(true);
            (result.inputs === undefined).should.equal(true);
        });
        it('matches fixture', async function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                krsProvider: 'dai',
                backupKey: 'test-backup-key',
                unspents,
            });
            const fixture = await (0, util_1.getFixture)(coin, 'recovery/formatBackupKeyRecovery-krs-dai', result);
            (0, util_1.shouldEqualJSON)(result, fixture);
        });
    });
    describe('full recovery (user + backup signatures)', function () {
        let psbt;
        let signingKeys;
        let unspents;
        before(function () {
            unspents = createTestUnspents();
            signingKeys = buildSigningKeys({ neuterUser: false, neuterBackup: false });
            psbt = (0, src_1.backupKeyRecoveryWithWalletUnspents)(coin.name, { walletKeys: defaultWasmWalletKeys, keys: signingKeys, recoveryDestination, feeRateSatVB }, unspents);
        });
        it('returns BackupKeyRecoveryTransansaction with finalized tx hex', function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                unspents,
            });
            result.should.have.property('transactionHex');
            result.should.have.property('inputs');
            const txBuf = Buffer.from(result.transactionHex, 'hex');
            (0, wasm_utxo_1.hasPsbtMagic)(txBuf).should.equal(false);
        });
        it('matches fixture', async function () {
            const result = (0, src_1.formatBackupKeyRecoveryResult)(coin, clonePsbt(psbt), {
                walletKeys: defaultWasmWalletKeys,
                keys: signingKeys,
                recoveryDestination,
                unspents,
            });
            const fixture = await (0, util_1.getFixture)(coin, 'recovery/formatBackupKeyRecovery-fullySigned', result);
            (0, util_1.shouldEqualJSON)(result, fixture);
        });
    });
});
