// @bitgo/abstract-utxo/dist/cjs/test/unit/recovery/mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WasmCrossChainRecoveryProvider = exports.MockRecoveryProvider = void 0;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const bch_1 = require("../../../src/impl/bch");
const bsv_1 = require("../../../src/impl/bsv");
const unspent_1 = require("../../../src/unspent");
class MockRecoveryProvider {
    constructor(unspents) {
        this.prevTxCache = {};
        this.unspents = unspents;
        this.unspents.forEach((u) => {
            if ('prevTx' in u) {
                const { txid } = (0, unspent_1.parseOutputId)(u.id);
                this.prevTxCache[txid] = u.prevTx.toString('hex');
            }
        });
    }
    async getAddressInfo(address) {
        const u = this.unspents.find((u) => u.address === address);
        return {
            txCount: u ? 1 : 0,
            balance: u ? Number(u.value) : 0,
        };
    }
    async getUnspentsForAddresses(addresses) {
        return this.unspents
            .filter((u) => addresses.includes(u.address))
            .map((u) => ({
            id: u.id,
            address: u.address,
            value: Number(u.value),
        }));
    }
    async getTransactionHex(txid) {
        return this.prevTxCache[txid];
    }
    getTransactionInputs(txid) {
        throw new Error(`not implemented`);
    }
    getTransactionIO(txid) {
        throw new Error(`not implemented`);
    }
}
exports.MockRecoveryProvider = MockRecoveryProvider;
/**
 * Cross-chain recovery mock provider using wasm-utxo only (no utxolib dependency).
 * Uses direct unspent data instead of parsing a utxolib transaction.
 */
class WasmCrossChainRecoveryProvider {
    constructor(coin, 
    /** The wallet unspent representing the output on the wrong chain */
    depositUnspent, 
    /** The addresses that were inputs to the original deposit transaction */
    inputAddresses) {
        this.coin = coin;
        this.depositUnspent = depositUnspent;
        this.inputAddresses = inputAddresses;
        this.addressFormat = this.coin instanceof bch_1.Bch && !(this.coin instanceof bsv_1.Bsv) ? 'cashaddr' : 'default';
        this.addressVersion = this.coin instanceof bch_1.Bch && !(this.coin instanceof bsv_1.Bsv) ? 'cashaddr' : 'base58';
    }
    async getUnspentsForAddresses(addresses) {
        // Format the deposit address for BCH-like coins
        let formattedAddress = this.depositUnspent.address;
        if (this.addressFormat === 'cashaddr') {
            formattedAddress = wasm_utxo_1.address.fromOutputScriptWithCoin(wasm_utxo_1.address.toOutputScriptWithCoin(this.depositUnspent.address, this.coin.name), this.coin.name, this.addressFormat);
            if (formattedAddress.includes(':')) {
                [, formattedAddress] = formattedAddress.split(':');
            }
        }
        return [
            {
                id: this.depositUnspent.id,
                address: formattedAddress,
                value: Number(this.depositUnspent.value),
                ...(this.coin.amountType === 'bigint' ? { valueString: this.depositUnspent.value.toString() } : {}),
            },
        ];
    }
    async getTransactionIO(txid) {
        // Format deposit address for output
        let outputAddress = this.depositUnspent.address;
        if (this.addressFormat === 'cashaddr') {
            outputAddress = wasm_utxo_1.address.fromOutputScriptWithCoin(wasm_utxo_1.address.toOutputScriptWithCoin(this.depositUnspent.address, this.coin.name), this.coin.name, this.addressFormat);
            if (outputAddress.includes(':')) {
                [, outputAddress] = outputAddress.split(':');
            }
        }
        return {
            inputs: this.inputAddresses.map((addr) => {
                let address = this.coin.canonicalAddress(addr, this.addressVersion);
                if (address.includes(':')) {
                    [, address] = address.split(':');
                }
                return { address };
            }),
            outputs: [{ address: outputAddress }],
        };
    }
    async getAddressInfo(address) {
        throw new Error('not implemented');
    }
    async getTransactionHex(txid) {
        throw new Error('not implemented');
    }
    getTransactionInputs(txid) {
        throw new Error('not implemented');
    }
}
exports.WasmCrossChainRecoveryProvider = WasmCrossChainRecoveryProvider;
