// @bitgo/sdk-coin-cronos/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GAS_LIMIT = exports.GAS_AMOUNT = exports.TESTNET_ADDRESS_PREFIX = exports.testnetContractAddressRegex = exports.testnetValidatorAddressRegex = exports.testnetAccountAddressRegex = exports.MAINNET_ADDRESS_PREFIX = exports.mainnetContractAddressRegex = exports.mainnetValidatorAddressRegex = exports.mainnetAccountAddressRegex = exports.validDenoms = void 0;
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const cosmosUtils = new abstract_cosmos_1.CosmosUtils();
exports.validDenoms = [
    'cro',
    'tcro',
    'basecro',
    'basetcro',
    ...cosmosUtils.getTokenDenomsUsingCoinFamily('cronos'),
];
exports.mainnetAccountAddressRegex = /^(cro)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.mainnetValidatorAddressRegex = /^(crocncl)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.mainnetContractAddressRegex = /^(cro)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]+)$/;
exports.MAINNET_ADDRESS_PREFIX = 'cro';
exports.testnetAccountAddressRegex = /^(tcro)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.testnetValidatorAddressRegex = /^(tcrocncl)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.testnetContractAddressRegex = /^(tcro)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]+)$/;
exports.TESTNET_ADDRESS_PREFIX = 'tcro';
exports.GAS_AMOUNT = '30000';
exports.GAS_LIMIT = 500000;
