// @bitgo/sdk-coin-xrp/dist/src/lib/accountSetBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountSetBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const iface_1 = require("./iface");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class AccountSetBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.AccountUpdate;
    }
    get xrpTransactionType() {
        return iface_1.XrpTransactionType.AccountSet;
    }
    setFlag(setFlag) {
        utils_1.default.validateAccountSetFlag(setFlag);
        this._setFlag = setFlag;
        return this;
    }
    messageKey(messageKey) {
        if (typeof messageKey !== 'string') {
            throw new sdk_core_1.BuildTransactionError('Invalid message key');
        }
        this._messageKey = messageKey;
        return this;
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        const { setFlag, messageKey } = tx.toJson();
        if (setFlag) {
            this.setFlag(setFlag);
        }
        if (messageKey) {
            this.messageKey(messageKey);
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        if (!this._sender) {
            throw new sdk_core_1.BuildTransactionError('Sender must be set before building the transaction');
        }
        const accountSetFields = {
            TransactionType: this.xrpTransactionType,
            Account: this._sender,
        };
        if (this._setFlag) {
            accountSetFields.SetFlag = this._setFlag;
        }
        if (this._messageKey) {
            accountSetFields.MessageKey = this._messageKey;
        }
        this._specificFields = accountSetFields;
        return await super.buildImplementation();
    }
}
exports.AccountSetBuilder = AccountSetBuilder;
