// @bitgo/sdk-coin-xrp/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REQUIRE_DESTINATION_TAG_FLAG = exports.MASTER_KEY_DEACTIVATION_FLAG = exports.USER_KEY_SETTING_FLAG = exports.VALID_ACCOUNT_SET_FLAGS = exports.MIN_SIGNER_QUORUM = exports.MIN_SIGNERS = exports.MAX_SIGNERS = void 0;
// https://xrpl.org/signerlistset.html
exports.MAX_SIGNERS = 32;
exports.MIN_SIGNERS = 1;
exports.MIN_SIGNER_QUORUM = 1;
// https://xrpl.org/accountset.html#accountset-flags
exports.VALID_ACCOUNT_SET_FLAGS = [
    5, // asfAccountTxnID
    16, // asfAllowTrustLineClawback
    10, // asfAuthorizedNFTokenMinter
    8, // asfDefaultRipple
    9, // asfDepositAuth
    4, // asfDisableMaster
    13, // asfDisallowIncomingCheck
    12, // asfDisallowIncomingNFTokenOffer
    14, // asfDisallowIncomingPayChan
    15, // asfDisallowIncomingTrustline
    3, // asfDisallowXRP
    7, // asfGlobalFreeze
    6, // asfNoFreeze
    2, // asfRequireAuth
    1, // asfRequireDest
];
// Global flags for bitgo address
exports.USER_KEY_SETTING_FLAG = 65536;
exports.MASTER_KEY_DEACTIVATION_FLAG = 1048576;
exports.REQUIRE_DESTINATION_TAG_FLAG = 131072;
