// @bitgo/sdk-coin-xrp/dist/src/lib/keyPair.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const secp256k1_1 = require("@bitgo/secp256k1");
const crypto_1 = require("crypto");
const xrpl = __importStar(require("xrpl"));
const utils_1 = __importDefault(require("./utils"));
const DEFAULT_SEED_SIZE_BYTES = 32;
/**
 * XRP keys and address management.
 */
class KeyPair extends sdk_core_1.Secp256k1ExtendedKeyPair {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param {KeyPairOptions} source Either a master seed, a private key, or a public key
     */
    constructor(source) {
        super(source);
        if (!source) {
            const seed = (0, crypto_1.randomBytes)(DEFAULT_SEED_SIZE_BYTES);
            this.hdNode = secp256k1_1.bip32.fromSeed(seed);
        }
        else if ((0, sdk_core_1.isSeed)(source)) {
            this.hdNode = secp256k1_1.bip32.fromSeed(source.seed);
        }
        else if ((0, sdk_core_1.isPrivateKey)(source)) {
            super.recordKeysFromPrivateKey(source.prv);
        }
        else if ((0, sdk_core_1.isPublicKey)(source)) {
            super.recordKeysFromPublicKey(source.pub);
        }
        else {
            throw new Error('Invalid key pair options');
        }
        if (this.hdNode) {
            this.keyPair = sdk_core_1.Secp256k1ExtendedKeyPair.toKeyPair(this.hdNode);
        }
    }
    /** @inheritdoc */
    getKeys() {
        return {
            pub: this.getPublicKey({ compressed: true }).toString('hex'),
            prv: this.getPrivateKey()?.toString('hex'),
        };
    }
    /** @inheritdoc */
    getAddress() {
        return xrpl.deriveAddress(this.getKeys().pub);
    }
    /**
     * Generates a signature for an arbitrary string with the current private key using keccak256
     * hashing algorithm. Throws if there is no private key.
     *
     * @param {string} message to produce a signature for
     * @returns {Buffer} The signature as a buffer
     */
    signMessage(message) {
        const messageToSign = Buffer.from(message).toString('hex');
        const { prv } = this.getKeys();
        if (!prv) {
            throw new Error('Missing private key');
        }
        const signature = utils_1.default.signString(messageToSign, prv);
        return Buffer.from(signature, 'hex');
    }
    /**
     * Verifies a message signature using the current public key.
     *
     * @param {string} message signed
     * @param {Buffer} signature to verify
     * @returns {boolean} True if the message was signed with the current key pair
     */
    verifySignature(message, signature) {
        const messageToVerify = Buffer.from(message).toString('hex');
        const pubKey = this.getKeys().pub;
        return utils_1.default.verifySignature(messageToVerify, pubKey, signature.toString('hex'));
    }
}
exports.KeyPair = KeyPair;
