// @bitgo/sdk-coin-xrp/dist/src/lib/tokenTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const iface_1 = require("./iface");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class TokenTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    get xrpTransactionType() {
        return iface_1.XrpTransactionType.Payment;
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        const { destination, amount, destinationTag } = tx.toJson();
        if (!destination) {
            throw new sdk_core_1.BuildTransactionError('Missing destination');
        }
        if (!amount) {
            throw new sdk_core_1.BuildTransactionError('Missing amount');
        }
        const normalizeAddress = utils_1.default.normalizeAddress({ address: destination, destinationTag });
        this.to(normalizeAddress);
        if (!utils_1.default.isIssuedCurrencyAmount(amount)) {
            throw new sdk_core_1.BuildTransactionError('Invalid Amount');
        }
        const amountBigNum = (0, bignumber_js_1.default)(amount.value).shiftedBy(this._coinConfig.decimalPlaces);
        this.amount(amountBigNum.toFixed());
    }
    /**
     *  Set the receiver address
     * @param {string} address - the address with optional destination tag
     * @returns {TransactionBuilder} This transaction builder
     */
    to(address) {
        const { address: xrpAddress, destinationTag } = utils_1.default.getAddressDetails(address);
        this._destination = xrpAddress;
        this._destinationTag = destinationTag;
        return this;
    }
    /**
     *  Set the amount to send
     * @param {string} amount - the amount sent
     * @returns {TransactionBuilder} This transaction builder
     */
    amount(amount) {
        if (typeof amount !== 'string') {
            throw new Error(`amount type ${typeof amount} must be a string`);
        }
        const amountBigNum = (0, bignumber_js_1.default)(amount);
        if (amountBigNum.lt(0)) {
            throw new Error(`amount ${amount} is not valid`);
        }
        const currency = utils_1.default.getXrpCurrencyFromTokenName(this._coinConfig.name);
        // Unlike most coins, XRP Token amounts are represented in decimal notation
        const value = amountBigNum.shiftedBy(-1 * this._coinConfig.decimalPlaces).toFixed();
        this._amount = {
            value: value,
            ...currency,
        };
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        if (!this._sender) {
            throw new sdk_core_1.BuildTransactionError('Sender must be set before building the transaction');
        }
        const transferFields = {
            TransactionType: this.xrpTransactionType,
            Account: this._sender,
            Destination: this._destination,
            Amount: this._amount,
        };
        if (typeof this._destinationTag === 'number') {
            transferFields.DestinationTag = this._destinationTag;
        }
        this._specificFields = transferFields;
        return await super.buildImplementation();
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
