// @bitgo/sdk-coin-xrp/dist/src/lib/transactionBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const _ = __importStar(require("lodash"));
const xrpl = __importStar(require("xrpl"));
const keyPair_1 = require("./keyPair");
const transaction_1 = require("./transaction");
const utils_1 = __importDefault(require("./utils"));
/**
 * XRP transaction builder.
 */
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._flags = 0;
        this.transaction = new transaction_1.Transaction(_coinConfig);
        this._keyPairs = [];
        this._signers = [];
    }
    /**
     * Set the transaction signature type to multi sig.
     **/
    setMultiSig() {
        this._isMultiSig = true;
        return this;
    }
    /**
     * Set the transaction signature type to single sig.
     **/
    setSingleSig() {
        this._isMultiSig = false;
        return this;
    }
    /**
     * Sets the sender of this transaction.
     *
     * @param {string} address the account that is sending this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    sender(address) {
        this.validateAddress({ address });
        this._sender = address;
        return this;
    }
    sequence(sequence) {
        if (typeof sequence !== 'number' || sequence < 0) {
            throw new Error(`sequence ${sequence} is not valid`);
        }
        this._sequence = sequence;
        return this;
    }
    fee(fee) {
        if (typeof fee !== 'string') {
            throw new Error(`fee type ${typeof fee} must be a string`);
        }
        const feeBigInt = BigInt(fee);
        if (feeBigInt < 0) {
            throw new Error(`fee ${fee} is not valid`);
        }
        this._fee = fee;
        return this;
    }
    flags(flags) {
        if (typeof flags !== 'number' || flags < 0) {
            throw new Error(`flags ${flags} is not valid`);
        }
        this._flags = flags;
        return this;
    }
    lastLedgerSequence(lastLedgerSequence) {
        if (typeof lastLedgerSequence !== 'number' || lastLedgerSequence < 0) {
            throw new Error(`lastLedgerSequence ${lastLedgerSequence} is not valid`);
        }
        this._lastLedgerSequence = lastLedgerSequence;
        return this;
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        this._transaction = tx;
        const txData = tx.toJson();
        if (!_.isUndefined(txData.isMultiSig)) {
            txData.isMultiSig ? this.setMultiSig() : this.setSingleSig();
        }
        this.sender(txData.from);
        if (txData.fee) {
            this.fee(txData.fee);
        }
        if (txData.sequence) {
            this.sequence(txData.sequence);
        }
        if (txData.lastLedgerSequence) {
            this.lastLedgerSequence(txData.lastLedgerSequence);
        }
        if (txData.flags) {
            this.flags(txData.flags);
        }
        this._signers = txData.signers || [];
        this._signingPubKey = txData.signingPubKey;
        this._txnSignature = txData.txnSignature;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = new transaction_1.Transaction(this._coinConfig);
        this.validateRawTransaction(rawTransaction);
        tx.fromRawTransaction(rawTransaction);
        this.initBuilder(tx);
        return this.transaction;
    }
    /** @inheritdoc */
    signImplementation(key) {
        this.validateKey(key);
        this.checkDuplicatedKeys(key);
        let prv = key.key;
        if (prv.startsWith('xprv')) {
            const rawPrv = (0, sdk_core_1.xprvToRawPrv)(prv);
            prv = new keyPair_1.KeyPair({ prv: rawPrv }).getKeys().prv;
        }
        const signer = new keyPair_1.KeyPair({ prv: prv });
        this._keyPairs.push(signer);
        return this.transaction;
    }
    checkDuplicatedKeys(key) {
        const keyPair = new keyPair_1.KeyPair({ prv: key.key }).getKeys();
        const keyPairPrv = keyPair.prv;
        this._keyPairs.forEach((kp) => {
            const prv = kp.getKeys().prv;
            if (prv.toLowerCase() === keyPairPrv.toLowerCase()) {
                throw new sdk_core_1.SigningError('Repeated sign');
            }
        });
        this._signers.forEach((signer) => {
            if (signer.Signer.SigningPubKey.toLowerCase() === keyPair.pub.toLowerCase()) {
                throw new sdk_core_1.SigningError('Repeated sign');
            }
        });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.xrpTransaction = this.buildXrpTransaction();
        this.transaction.setTransactionType(this.transactionType);
        this.transaction.loadInputsAndOutputs();
        if (this._keyPairs.length > 0) {
            this.validateIsMultiSig();
            this.transaction.setMultiSigValue(this._isMultiSig);
            this.transaction.sign(this._keyPairs);
        }
        return this.transaction;
    }
    buildXrpTransaction() {
        const commonFields = {
            Account: this._sender,
            Fee: this._fee,
            Sequence: this._sequence,
            Flags: this._flags,
        };
        if (this._signingPubKey) {
            commonFields.SigningPubKey = this._signingPubKey;
        }
        if (this._txnSignature) {
            commonFields.TxnSignature = this._txnSignature;
        }
        if (this._signers.length > 0) {
            commonFields.Signers = this._signers;
        }
        if (this._lastLedgerSequence) {
            commonFields.LastLedgerSequence = this._lastLedgerSequence;
        }
        const tx = Object.assign(commonFields, this._specificFields);
        xrpl.validate(tx);
        return tx;
    }
    validateKey(key) {
        let keyPair;
        try {
            keyPair = new keyPair_1.KeyPair({ prv: key.key });
        }
        catch {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
        if (!keyPair.getKeys().prv) {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
    }
    validateIsMultiSig() {
        if (_.isUndefined(this._isMultiSig)) {
            throw new sdk_core_1.BuildTransactionError('Signature type is not defined. Please call setMultiSig or setSingleSig.');
        }
    }
    /** @inheritdoc */
    validateTransaction() {
        if (this._sender === undefined) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: missing sender');
        }
        if (this._fee === undefined) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: missing fee');
        }
        if (this._sequence === undefined) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: missing sequence');
        }
    }
    validateAddress(address) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address ' + address.address);
        }
    }
    validateValue(value) {
        if (value.isLessThan(0)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be less than zero');
        }
    }
    validateRawTransaction(rawTransaction) {
        utils_1.default.validateRawTransaction(rawTransaction);
    }
    get transaction() {
        return this._transaction;
    }
    set transaction(transaction) {
        this._transaction = transaction;
    }
}
exports.TransactionBuilder = TransactionBuilder;
