// @bitgo/sdk-coin-xrp/dist/src/lib/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const lodash_1 = __importDefault(require("lodash"));
const xrpl = __importStar(require("xrpl"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("./utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const iface_1 = require("./iface");
/**
 * XRP transaction.
 */
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
    }
    get xrpTransaction() {
        return this._xrpTransaction;
    }
    set xrpTransaction(tx) {
        this._xrpTransaction = tx;
    }
    canSign(key) {
        if (!utils_1.default.isValidPrivateKey(key.key)) {
            return false;
        }
        return true;
    }
    toJson() {
        if (!this._xrpTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        const txData = {
            from: this._xrpTransaction.Account,
            isMultiSig: this._isMultiSig,
            transactionType: this._xrpTransaction.TransactionType,
            id: this._id,
            fee: this._xrpTransaction.Fee,
            sequence: this._xrpTransaction.Sequence,
            lastLedgerSequence: this._xrpTransaction.LastLedgerSequence,
            flags: this._xrpTransaction.Flags,
            signingPubKey: this._xrpTransaction.SigningPubKey,
            signers: this._xrpTransaction.Signers,
            txnSignature: this._xrpTransaction.TxnSignature,
        };
        if (this._xrpTransaction.SigningPubKey === '' && !lodash_1.default.isEmpty(this._xrpTransaction.Signers)) {
            txData.isMultiSig = true;
        }
        if (this._xrpTransaction.SigningPubKey && utils_1.default.isValidPublicKey(this._xrpTransaction.SigningPubKey)) {
            txData.isMultiSig = false;
        }
        switch (this._xrpTransaction.TransactionType) {
            case iface_1.XrpTransactionType.Payment:
                txData.destination = this._xrpTransaction.Destination;
                txData.destinationTag = this._xrpTransaction.DestinationTag;
                if (typeof this._xrpTransaction.Amount === 'string' ||
                    utils_1.default.isIssuedCurrencyAmount(this._xrpTransaction.Amount)) {
                    txData.amount = this._xrpTransaction.Amount;
                }
                else {
                    throw new sdk_core_1.InvalidTransactionError('Invalid amount');
                }
                return txData;
            case iface_1.XrpTransactionType.AccountSet:
                txData.setFlag = this._xrpTransaction.SetFlag;
                txData.messageKey = this._xrpTransaction.MessageKey;
                return txData;
            case iface_1.XrpTransactionType.SignerListSet:
                txData.signerQuorum = this._xrpTransaction.SignerQuorum;
                txData.signerEntries = this._xrpTransaction.SignerEntries;
                return txData;
            case iface_1.XrpTransactionType.TrustSet:
                txData.amount = this._xrpTransaction.LimitAmount;
                return txData;
            default:
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction type');
        }
    }
    getSignablePayload() {
        if (!this._xrpTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        return lodash_1.default.omit(this._xrpTransaction, ['TxnSignature', 'Signers', 'SigningPubKey']);
    }
    sign(keyPair) {
        if (!this._xrpTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        if (!this._xrpTransaction.Fee) {
            throw new sdk_core_1.InvalidTransactionError('Missing fee');
        }
        if (!this._xrpTransaction.Sequence) {
            throw new sdk_core_1.InvalidTransactionError('Missing sequence');
        }
        if (!this._xrpTransaction.Flags) {
            throw new sdk_core_1.InvalidTransactionError('Missing flags');
        }
        if (lodash_1.default.isEmpty(keyPair)) {
            return;
        }
        const keyPairs = keyPair instanceof Array ? keyPair : [keyPair];
        for (const kp of keyPairs) {
            const { pub, prv } = kp.getKeys();
            if (!prv) {
                throw new sdk_core_1.SigningError('Missing private key');
            }
            if (this._isMultiSig === false && this._xrpTransaction.TxnSignature) {
                throw new sdk_core_1.SigningError('Transaction has already been signed');
            }
            const signablePayload = this.getSignablePayload();
            const xrpWallet = new xrpl.Wallet(pub, prv);
            const signedTx = xrpWallet.sign(signablePayload, this._isMultiSig);
            const xrpSignedTx = xrpl.decode(signedTx.tx_blob);
            xrpl.validate(xrpSignedTx);
            if (this._isMultiSig === false) {
                xrpWallet.verifyTransaction(signedTx.tx_blob);
                this._xrpTransaction = xrpSignedTx;
                this._id = signedTx.hash;
            }
            if (this._isMultiSig === true) {
                if (!xrpSignedTx.Signers || !lodash_1.default.isArray(xrpSignedTx.Signers)) {
                    throw new sdk_core_1.SigningError('Missing or invalid signers');
                }
                const sortedSigners = this.concatAndSortSigners(this._xrpTransaction.Signers || [], xrpSignedTx.Signers);
                this._xrpTransaction = xrpSignedTx;
                this._xrpTransaction.Signers = sortedSigners;
                this._id = this.calculateIdFromRawTx(xrpl.encode(this._xrpTransaction));
            }
        }
    }
    /** @inheritdoc */
    toBroadcastFormat() {
        if (!this._xrpTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        return xrpl.encode(this._xrpTransaction);
    }
    explainTransaction() {
        switch (this._xrpTransaction.TransactionType) {
            case iface_1.XrpTransactionType.Payment:
                return this.explainPaymentTransaction();
            case iface_1.XrpTransactionType.AccountSet:
                return this.explainAccountSetTransaction();
            case iface_1.XrpTransactionType.SignerListSet:
                return this.explainSignerListSetTransaction();
            default:
                throw new Error('Unsupported transaction type');
        }
    }
    explainPaymentTransaction() {
        const tx = this._xrpTransaction;
        const address = utils_1.default.normalizeAddress({ address: tx.Destination, destinationTag: tx.DestinationTag });
        const amount = lodash_1.default.isString(tx.Amount) ? tx.Amount : 0;
        return {
            displayOrder: ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee'],
            id: this._id,
            changeOutputs: [],
            outputAmount: amount,
            changeAmount: 0,
            outputs: [
                {
                    address,
                    amount,
                },
            ],
            fee: {
                fee: tx.Fee,
                feeRate: undefined,
            },
        };
    }
    explainAccountSetTransaction() {
        const tx = this._xrpTransaction;
        return {
            displayOrder: ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'accountSet'],
            id: this._id,
            changeOutputs: [],
            outputAmount: 0,
            changeAmount: 0,
            outputs: [],
            fee: {
                fee: tx.Fee,
                feeRate: undefined,
            },
            accountSet: {
                messageKey: tx.MessageKey,
                setFlag: tx.SetFlag,
            },
        };
    }
    explainSignerListSetTransaction() {
        const tx = this._xrpTransaction;
        return {
            displayOrder: ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'signerListSet'],
            id: this._id,
            changeOutputs: [],
            outputAmount: 0,
            changeAmount: 0,
            outputs: [],
            fee: {
                fee: tx.Fee,
                feeRate: undefined,
            },
            signerListSet: {
                signerQuorum: tx.SignerQuorum,
                signerEntries: tx.SignerEntries,
            },
        };
    }
    calculateIdFromRawTx(rawTransaction) {
        let id;
        // hashes ids are different for signed and unsigned tx
        // first we try to get the hash id as if it is signed, will throw if its not
        try {
            id = xrpl.hashes.hashSignedTx(rawTransaction);
        }
        catch (e) {
            id = xrpl.hashes.hashTx(rawTransaction);
        }
        return id;
    }
    /**
     * Set the transaction type.
     *
     * @param {TransactionType} transactionType The transaction type to be set.
     */
    setTransactionType(transactionType) {
        this._type = transactionType;
    }
    setMultiSigValue(isMultiSig) {
        this._isMultiSig = isMultiSig;
    }
    /**
     * Sets this transaction payload
     *
     * @param rawTransaction
     */
    fromRawTransaction(rawTransaction) {
        let txHex = rawTransaction;
        if (!utils_1.default.isValidHex(rawTransaction)) {
            try {
                txHex = xrpl.encode(JSON.parse(rawTransaction));
            }
            catch (e) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction');
            }
        }
        utils_1.default.validateRawTransaction(txHex);
        this._xrpTransaction = xrpl.decode(txHex);
        if (!iface_1.XrpTransactionType[this._xrpTransaction.TransactionType]) {
            throw new sdk_core_1.InvalidTransactionError('Unsupported transaction type, got: ' + this._xrpTransaction.TransactionType);
        }
        if (this._xrpTransaction.SigningPubKey && this._xrpTransaction.SigningPubKey !== '') {
            this._isMultiSig = false;
        }
        if (this._xrpTransaction.SigningPubKey === '' &&
            this._xrpTransaction.Signers &&
            this._xrpTransaction.Signers.length > 0) {
            this._isMultiSig = true;
        }
        this._id = this.calculateIdFromRawTx(txHex);
        switch (this._xrpTransaction.TransactionType) {
            case iface_1.XrpTransactionType.SignerListSet:
                this.setTransactionType(sdk_core_1.TransactionType.WalletInitialization);
                break;
            case iface_1.XrpTransactionType.AccountSet:
                this.setTransactionType(sdk_core_1.TransactionType.AccountUpdate);
                break;
            case iface_1.XrpTransactionType.Payment:
                if (utils_1.default.isIssuedCurrencyAmount(this._xrpTransaction.Amount)) {
                    this.setTransactionType(sdk_core_1.TransactionType.SendToken);
                }
                else {
                    this.setTransactionType(sdk_core_1.TransactionType.Send);
                }
                break;
            case iface_1.XrpTransactionType.TrustSet:
                this.setTransactionType(sdk_core_1.TransactionType.TrustLine);
                break;
        }
        this.loadInputsAndOutputs();
    }
    /**
     * Load the input and output data on this transaction.
     */
    loadInputsAndOutputs() {
        if (!this._xrpTransaction) {
            return;
        }
        if (this._xrpTransaction.TransactionType === iface_1.XrpTransactionType.Payment) {
            let value;
            const { Account, Destination, Amount, DestinationTag } = this._xrpTransaction;
            if (typeof Amount === 'string') {
                value = Amount;
            }
            else {
                value = Amount.value;
            }
            const coin = this._coinConfig.name;
            this.inputs.push({
                address: Account,
                value,
                coin,
            });
            this.outputs.push({
                address: utils_1.default.normalizeAddress({ address: Destination, destinationTag: DestinationTag }),
                value,
                coin,
            });
        }
    }
    /**
     * Groups and sorts the signers by account.
     * @param {Signer[]}signers1 - The first set of signers.
     * @param {Signer[]}signers2 - The second set of signers.
     * @returns The grouped and sorted signers.
     **/
    concatAndSortSigners(signers1, signers2) {
        return signers1
            .concat(signers2)
            .sort((signer1, signer2) => this.compareSignersByAccount(signer1.Signer.Account, signer2.Signer.Account));
    }
    /**
     * If presented in binary form, the Signers array must be sorted based on
     * the numeric value of the signer addresses, with the lowest value first.
     * (If submitted as JSON, the submit_multisigned method handles this automatically.)
     * https://xrpl.org/multi-signing.html.
     *
     * @param left - A Signer to compare with.
     * @param right - A second Signer to compare with.
     * @returns 1 if left \> right, 0 if left = right, -1 if left \< right, and null if left or right are NaN.
     */
    compareSignersByAccount(address1, address2) {
        const addressBN1 = this.addressToBigNumber(address1);
        const addressBN2 = this.addressToBigNumber(address2);
        return addressBN1.comparedTo(addressBN2);
    }
    addressToBigNumber(address) {
        const hex = Buffer.from(xrpl.decodeAccountID(address)).toString('hex');
        const numberOfBitsInHex = 16;
        return new bignumber_js_1.default(hex, numberOfBitsInHex);
    }
}
exports.Transaction = Transaction;
