// @bitgo/sdk-coin-xrp/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const xrpl_1 = __importDefault(require("xrpl"));
const accountSetBuilder_1 = require("./accountSetBuilder");
const tokenTransferBuilder_1 = require("./tokenTransferBuilder");
const transaction_1 = require("./transaction");
const transferBuilder_1 = require("./transferBuilder");
const trustsetBuilder_1 = require("./trustsetBuilder");
const utils_1 = __importDefault(require("./utils"));
const walletInitializationBuilder_1 = require("./walletInitializationBuilder");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    from(raw) {
        let txHex = raw;
        if (!utils_1.default.isValidHex(raw)) {
            try {
                txHex = xrpl_1.default.encode(JSON.parse(raw));
            }
            catch (e) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction');
            }
        }
        const tx = this.parseTransaction(txHex);
        try {
            switch (tx.type) {
                case sdk_core_1.TransactionType.AccountUpdate:
                    return this.getAccountUpdateBuilder(tx);
                case sdk_core_1.TransactionType.Send:
                    return this.getTransferBuilder(tx);
                case sdk_core_1.TransactionType.WalletInitialization:
                    return this.getWalletInitializationBuilder(tx);
                case sdk_core_1.TransactionType.SendToken:
                    return this.getTokenTransferBuilder(tx);
                case sdk_core_1.TransactionType.TrustLine:
                    return this.getTrustSetBuilder(tx);
                default:
                    throw new sdk_core_1.InvalidTransactionError('Invalid transaction');
            }
        }
        catch (e) {
            throw e;
        }
    }
    /** @inheritdoc */
    getWalletInitializationBuilder(tx) {
        return this.initializeBuilder(tx, new walletInitializationBuilder_1.WalletInitializationBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getTransferBuilder(tx) {
        return this.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getAccountUpdateBuilder(tx) {
        return this.initializeBuilder(tx, new accountSetBuilder_1.AccountSetBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getTokenTransferBuilder(tx) {
        return this.initializeBuilder(tx, new tokenTransferBuilder_1.TokenTransferBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getTrustSetBuilder(tx) {
        return this.initializeBuilder(tx, new trustsetBuilder_1.TrustSetBuilder(this._coinConfig));
    }
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
    /** Parse the transaction from a raw transaction
     *
     * @param {string} rawTransaction - the raw tx
     * @returns {Transaction} parsedtransaction
     */
    parseTransaction(rawTransaction) {
        const tx = new transaction_1.Transaction(this._coinConfig);
        tx.fromRawTransaction(rawTransaction);
        return tx;
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
