// @bitgo/sdk-coin-xrp/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const querystring = __importStar(require("querystring"));
const rippleKeypairs = __importStar(require("ripple-keypairs"));
const url = __importStar(require("url"));
const xrpl = __importStar(require("xrpl"));
const constants_1 = require("./constants");
const keyPair_1 = require("./keyPair");
const assert_1 = __importDefault(require("assert"));
class Utils {
    isValidAddress(address) {
        try {
            const addressDetails = this.getAddressDetails(address);
            return address === this.normalizeAddress(addressDetails);
        }
        catch (e) {
            return false;
        }
    }
    isValidTransactionId(txId) {
        return this.isValidHex(txId);
    }
    isValidPublicKey(key) {
        try {
            new keyPair_1.KeyPair({ pub: key });
            return true;
        }
        catch {
            return false;
        }
    }
    isValidPrivateKey(key) {
        try {
            new keyPair_1.KeyPair({ prv: key });
            return true;
        }
        catch {
            return false;
        }
    }
    isValidSignature(signature) {
        return this.isValidHex(signature);
    }
    isValidBlockId(hash) {
        return this.isValidHex(hash);
    }
    isValidHex(hex) {
        return /^([a-fA-F0-9])+$/.test(hex);
    }
    /**
     * Parse an address string into address and destination tag
     */
    getAddressDetails(address) {
        const destinationDetails = url.parse(address);
        const destinationAddress = destinationDetails.pathname;
        if (!destinationAddress || !xrpl.isValidClassicAddress(destinationAddress)) {
            throw new sdk_core_1.InvalidAddressError(`destination address "${destinationAddress}" is not valid`);
        }
        // there are no other properties like destination tags
        if (destinationDetails.pathname === address) {
            return {
                address: address,
                destinationTag: undefined,
            };
        }
        if (!destinationDetails.query) {
            throw new sdk_core_1.InvalidAddressError('no query params present');
        }
        const queryDetails = querystring.parse(destinationDetails.query);
        if (!queryDetails.dt) {
            // if there are more properties, the query details need to contain the destination tag property.
            throw new sdk_core_1.InvalidAddressError('destination tag missing');
        }
        if (Array.isArray(queryDetails.dt)) {
            // if queryDetails.dt is an array, that means dt was given multiple times, which is not valid
            throw new sdk_core_1.InvalidAddressError(`destination tag can appear at most once, but ${queryDetails.dt.length} destination tags were found`);
        }
        const parsedTag = parseInt(queryDetails.dt, 10);
        if (!Number.isSafeInteger(parsedTag)) {
            throw new sdk_core_1.InvalidAddressError('invalid destination tag');
        }
        if (parsedTag > 0xffffffff || parsedTag < 0) {
            throw new sdk_core_1.InvalidAddressError('destination tag out of range');
        }
        return {
            address: destinationAddress,
            destinationTag: parsedTag,
        };
    }
    /**
     * Construct a full, normalized address from an address and destination tag
     */
    normalizeAddress({ address, destinationTag }) {
        if (typeof address !== 'string') {
            throw new sdk_core_1.InvalidAddressError('invalid address, expected string');
        }
        if (typeof destinationTag === 'undefined' || destinationTag === null) {
            return address;
        }
        if (!Number.isInteger(destinationTag)) {
            throw new sdk_core_1.InvalidAddressError('invalid destination tag, expected integer');
        }
        if (destinationTag > 0xffffffff || destinationTag < 0) {
            throw new sdk_core_1.InvalidAddressError('destination tag out of range');
        }
        return `${address}?dt=${destinationTag}`;
    }
    /**
     * @param message hex encoded string
     * @param privateKey
     * return hex encoded signature string, throws if any of the inputs are invalid
     */
    signString(message, privateKey) {
        if (!this.isValidHex(message)) {
            throw new sdk_core_1.UtilsError('message must be a hex encoded string');
        }
        if (!this.isValidPrivateKey(privateKey)) {
            throw new sdk_core_1.UtilsError('invalid private key');
        }
        return rippleKeypairs.sign(message, privateKey);
    }
    /**
     * @param message hex encoded string
     * @param signature hex encooded signature string
     * @param publicKey
     * return boolean, throws if any of the inputs are invalid
     */
    verifySignature(message, signature, publicKey) {
        if (!this.isValidHex(message)) {
            throw new sdk_core_1.UtilsError('message must be a hex encoded string');
        }
        if (!this.isValidSignature(signature)) {
            throw new sdk_core_1.UtilsError('invalid signature');
        }
        if (!this.isValidPublicKey(publicKey)) {
            throw new sdk_core_1.UtilsError('invalid public key');
        }
        try {
            return rippleKeypairs.verify(message, signature, publicKey);
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Check the raw transaction has a valid format in the blockchain context, throw otherwise.
     *
     * @param {string} rawTransaction - Transaction in hex string format
     */
    validateRawTransaction(rawTransaction) {
        if (!rawTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Invalid raw transaction: Undefined');
        }
        if (!this.isValidHex(rawTransaction)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid raw transaction: Hex string expected');
        }
        if (!this.isValidRawTransaction(rawTransaction)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid raw transaction');
        }
    }
    /**
     * Checks if raw transaction can be deserialized
     *
     * @param {string} rawTransaction - transaction in base64 string format
     * @returns {boolean} - the validation result
     */
    isValidRawTransaction(rawTransaction) {
        try {
            const jsonTx = xrpl.decode(rawTransaction);
            xrpl.validate(jsonTx);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    validateAccountSetFlag(setFlag) {
        if (typeof setFlag !== 'number') {
            throw new sdk_core_1.UtilsError(`setFlag ${setFlag} is not valid`);
        }
        if (!constants_1.VALID_ACCOUNT_SET_FLAGS.includes(setFlag)) {
            throw new sdk_core_1.UtilsError(`setFlag ${setFlag} is not a valid account set flag`);
        }
    }
    validateSigner(signer) {
        if (!signer.address) {
            throw new sdk_core_1.UtilsError('signer must have an address');
        }
        if (!this.isValidAddress(signer.address)) {
            throw new sdk_core_1.UtilsError(`signer address ${signer.address} is invalid`);
        }
        if (typeof signer.weight !== 'number' || signer.weight < 0) {
            throw new sdk_core_1.UtilsError(`signer weight ${signer.weight} is not valid`);
        }
    }
    /**
     * Determines if the provided `amount` is for a token payment
     */
    isIssuedCurrencyAmount(amount) {
        return (!!amount &&
            typeof amount === 'object' &&
            typeof amount.currency === 'string' &&
            typeof amount.issuer === 'string' &&
            typeof amount.value === 'string');
    }
    /**
     * Get the associated XRP Currency details from token name. Throws an error if token is unsupported
     * @param {string} tokenName - The token name
     */
    getXrpCurrencyFromTokenName(tokenName) {
        if (!statics_1.coins.has(tokenName)) {
            throw new sdk_core_1.UnsupportedTokenError(`${tokenName} is not supported`);
        }
        const token = statics_1.coins.get(tokenName);
        if (!token.isToken || !(token instanceof statics_1.XrpCoin)) {
            throw new sdk_core_1.UnsupportedTokenError(`${tokenName} is not an XRP token`);
        }
        return {
            currency: token.currencyCode,
            issuer: token.issuerAddress,
        };
    }
    /**
     * Decodes a serialized XRPL transaction.
     *
     * @param {string} txHex - The serialized transaction in hex.
     * @returns {Object} - Decoded transaction object.
     * @throws {Error} - If decoding fails or input is invalid.
     */
    decodeTransaction(txHex) {
        if (typeof txHex !== 'string' || txHex.trim() === '') {
            throw new Error('Invalid transaction hex. Expected a non-empty string.');
        }
        try {
            return xrpl.decode(txHex);
        }
        catch (error) {
            throw new Error(`Failed to decode transaction: ${error.message}`);
        }
    }
    /**
     * Get the statics coin object matching a given Xrp token issuer address and currency code if it exists
     *
     * @param issuerAddress The token issuer address to match against
     * @param currencyCode The token currency code to match against
     * @returns statics BaseCoin object for the matching token
     */
    getXrpToken(issuerAddress, currencyCode) {
        const tokens = statics_1.coins.filter((coin) => {
            if (coin instanceof statics_1.XrpCoin) {
                return coin.issuerAddress === issuerAddress && coin.currencyCode === currencyCode;
            }
            return false;
        });
        const tokensArray = tokens.map((token) => token);
        if (tokensArray.length >= 1) {
            // there should never be two tokens with the same issuer address and currency code, so we assert that here
            (0, assert_1.default)(tokensArray.length === 1);
            return tokensArray[0];
        }
        return undefined;
    }
}
const utils = new Utils();
exports.default = utils;
