// @bitgo/sdk-coin-xrp/dist/src/lib/walletInitializationBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletInitializationBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class WalletInitializationBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._signerEntries = [];
    }
    get transactionType() {
        return sdk_core_1.TransactionType.WalletInitialization;
    }
    get xrpTransactionType() {
        return 'SignerListSet';
    }
    signerQuorum(quorum) {
        if (typeof quorum !== 'number' || quorum < 1) {
            throw new Error(`quorum must be a valid number greater than 0, got: ${quorum}`);
        }
        this._signerQuorum = quorum;
        return this;
    }
    signer(signer) {
        if (this._signerEntries.length > constants_1.MAX_SIGNERS) {
            throw new sdk_core_1.BuildTransactionError(`Cannot have more than ${constants_1.MAX_SIGNERS} signers`);
        }
        utils_1.default.validateSigner(signer);
        this._signerEntries.push({
            SignerEntry: { Account: signer.address, SignerWeight: signer.weight },
        });
        return this;
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        const { signerEntries, signerQuorum } = tx.toJson();
        if (signerQuorum) {
            this.signerQuorum(signerQuorum);
        }
        if (signerEntries?.length) {
            signerEntries.forEach((signer) => this.signer({
                address: signer.SignerEntry.Account,
                weight: signer.SignerEntry.SignerWeight,
            }));
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        if (!this._sender) {
            throw new sdk_core_1.BuildTransactionError('Sender must be set before building the transaction');
        }
        if (!this._signerQuorum || this._signerQuorum < constants_1.MIN_SIGNER_QUORUM) {
            throw new sdk_core_1.BuildTransactionError('Signer quorum must be set before building the transaction');
        }
        if (!this._signerEntries) {
            throw new sdk_core_1.BuildTransactionError('Signers must be set before building the transaction');
        }
        if (this._signerEntries.length < constants_1.MIN_SIGNERS || this._signerEntries.length > constants_1.MAX_SIGNERS) {
            throw new sdk_core_1.BuildTransactionError(`Signers must be between ${constants_1.MIN_SIGNERS} and ${constants_1.MAX_SIGNERS}`);
        }
        this._specificFields = {
            TransactionType: this.xrpTransactionType,
            Account: this._sender,
            SignerQuorum: this._signerQuorum,
            SignerEntries: this._signerEntries,
        };
        return await super.buildImplementation();
    }
}
exports.WalletInitializationBuilder = WalletInitializationBuilder;
