// @bitgo/public-types/dist/src/schema/transactionRequest/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const transactionState_1 = require("./transactionState");
const unsignedTransaction_1 = require("./unsignedTransaction");
const signatureShare_1 = require("./signatureShare");
const commitmentShare_1 = require("./commitmentShare");
exports.Transaction = t.intersection([
    t.type({
        state: transactionState_1.TransactionState,
        unsignedTx: unsignedTransaction_1.UnsignedTransaction,
    }),
    t.partial({
        signatureShares: t.array(signatureShare_1.SignatureShare),
        commitmentShares: t.array(commitmentShare_1.CommitmentShare),
        txHash: t.string,
        updatedDate: io_ts_types_1.DateFromISOString,
        createdDate: io_ts_types_1.DateFromISOString,
    }),
]);
