// @bitgo/public-types/dist/src/schema/transactionRequest/transactionRequestState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PendingTranscationRequestStates = exports.TransactionRequestState = void 0;
const t = __importStar(require("io-ts"));
exports.TransactionRequestState = t.keyof({
    pendingApproval: 1,
    canceled: 1,
    rejected: 1,
    initialized: 1,
    pendingDelivery: 1,
    delivered: 1,
    pendingUserSignature: 1,
    pendingUserCommitment: 1,
    pendingUserRShare: 1,
    pendingUserGShare: 1,
    readyToSend: 1,
    signed: 1,
    failed: 1,
}, "TransactionRequestState");
exports.PendingTranscationRequestStates = [
    "pendingApproval",
    "pendingDelivery",
    "pendingUserSignature",
    "pendingUserCommitment",
    "pendingUserRShare",
    "pendingUserGShare",
    "readyToSend",
];
