// @bitgo/public-types/dist/src/schema/transactionRequest/transactionRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionRequest = exports.TransactionRequestFull = exports.TransactionRequestLite = exports.TransactionRequestBase = exports.TransactionRequestIdNonEmptyString = exports.TransactionRequestId = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const wallet_1 = require("../wallet");
const walletTypes_1 = require("../wallet/walletTypes");
const enterprise_1 = require("../enterprise");
const _1 = require(".");
const transactionRequestApiVersion_1 = require("./transactionRequestApiVersion");
const commitmentShare_1 = require("./commitmentShare");
const signatureShare_1 = require("./signatureShare");
const unsignedTransaction_1 = require("./unsignedTransaction");
const message_1 = require("./message");
const transaction_1 = require("./transaction");
exports.TransactionRequestId = t.string;
exports.TransactionRequestIdNonEmptyString = io_ts_types_1.NonEmptyString;
exports.TransactionRequestBase = t.intersection([
    t.type({
        txRequestId: exports.TransactionRequestId,
        walletId: wallet_1.WalletIdString,
        version: t.number,
        state: _1.TransactionRequestState,
        date: io_ts_types_1.DateFromISOString,
        createdDate: io_ts_types_1.DateFromISOString,
        userId: t.string,
        initiatedBy: t.string,
        updatedBy: t.string,
        intents: t.array(t.unknown),
        latest: t.boolean,
    }),
    t.partial({
        apiVersion: transactionRequestApiVersion_1.TransactionRequestApiVersion,
        idempotencyKey: t.string,
        walletType: walletTypes_1.WalletTypes,
        enterpriseId: enterprise_1.EnterpriseIdString,
        intent: t.unknown,
        pendingApprovalId: t.string,
        policiesChecked: t.boolean,
        isCanceled: t.boolean,
    }),
]);
exports.TransactionRequestLite = t.intersection([
    exports.TransactionRequestBase,
    t.partial({
        intent: t.union([_1.TransactionIntent, t.unknown]),
        unsignedTxs: t.array(unsignedTransaction_1.UnsignedTransaction),
        signatureShares: t.array(signatureShare_1.SignatureShare),
        commitmentShares: t.array(commitmentShare_1.CommitmentShare),
        txHashes: t.array(t.string),
    }),
]);
exports.TransactionRequestFull = t.intersection([
    exports.TransactionRequestBase,
    t.partial({
        transactions: t.array(transaction_1.Transaction),
        messages: t.array(message_1.Message),
    }),
]);
exports.TransactionRequest = t.union([
    exports.TransactionRequestLite,
    exports.TransactionRequestFull,
]);
