// @bitgo/public-types/dist/src/schema/transactionRequest/unsignedTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsignedTransaction = void 0;
const t = __importStar(require("io-ts"));
const Input = t.intersection([
    t.type({
        valueString: t.string,
    }),
    t.partial({
        _id: t.string,
        address: t.string,
        chain: t.number,
        derivationIndex: t.number,
        index: t.number,
        value: t.union([t.number, t.bigint]),
    }),
]);
const Output = t.intersection([
    t.type({
        valueString: t.string,
    }),
    t.partial({
        address: t.string,
        change: t.boolean,
        coinName: t.string,
        data: t.string,
        isPayGo: t.boolean,
        value: t.union([t.number, t.bigint]),
        wallet: t.string,
        walletV1: t.string,
        baseAddress: t.string,
        enterprise: t.string,
    }),
]);
const ParsedTx = t.intersection([
    t.type({
        inputs: t.array(Input),
        minerFee: t.union([t.string, t.number]),
        outputs: t.array(Output),
        spendAmount: t.union([t.string, t.number]),
    }),
    t.partial({
        gasPrice: t.number,
        hasUnvalidatedData: t.boolean,
        payGoFee: t.union([t.string, t.number]),
        spendAmounts: t.array(t.UnknownRecord),
        type: t.string,
    }),
]);
exports.UnsignedTransaction = t.intersection([
    t.type({
        parsedTx: ParsedTx,
        serializedTxHex: t.string,
        signableHex: t.string,
    }),
    t.partial({
        derivationPath: t.string,
        coinSpecific: t.intersection([
            t.UnknownRecord,
            t.partial({
                nonceAddress: t.string,
                stakingAddress: t.string,
                signature: t.string,
            }),
        ]),
        feeInfo: t.type({
            fee: t.union([t.string, t.number]),
            feeString: t.string,
        }),
    }),
]);
