// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/addressInitializationBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressInitializationBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const abstract_eth_1 = require("@bitgo/abstract-eth");
const sdk_core_1 = require("@bitgo/sdk-core");
const ethereumjs_util_1 = require("ethereumjs-util");
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("../utils"));
class AddressInitializationBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new AddressInitializationBuilder instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * Initializes the builder with an existing AddressInitializationTransaction.
     *
     * @param {AddressInitializationTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the address initialization transaction instance.
     *
     * @returns {AddressInitializationTransaction} The address initialization transaction
     */
    get addressInitializationTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for address initialization.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.AddressInitialization;
    }
    /**
     * Validates the transaction clauses for address initialization.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For address init transactions, value must be exactly '0x0'
            if (clause.value !== 0) {
                return false;
            }
            const { baseAddress, addressCreationSalt, feeAddress } = (0, abstract_eth_1.decodeForwarderCreationData)(clause.data);
            if (!utils_1.default.isValidAddress(baseAddress) || !utils_1.default.isValidAddress(feeAddress)) {
                return false;
            }
            if (!addressCreationSalt) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the base address for this forwarder init tx.
     *
     * @param {string} address - The base address to be set for the forwarder address
     * @returns {AddressInitializationBuilder} This transaction builder
     */
    baseAddress(address) {
        this.validateAddress({ address });
        this.addressInitializationTransaction.baseAddress = address;
        return this;
    }
    /**
     * Sets the fee address for this forwarder init tx.
     *
     * @param {string} address - The fee address to be set for the forwarder address
     * @returns {AddressInitializationBuilder} This transaction builder
     */
    feeAddress(address) {
        this.validateAddress({ address });
        this.addressInitializationTransaction.feeAddress = address;
        return this;
    }
    /**
     * Sets the salt value for the forwarder address creation.
     *
     * @param {string} salt - The salt value to use for address creation
     * @returns {AddressInitializationBuilder} This transaction builder
     */
    salt(salt) {
        this.addressInitializationTransaction.salt = salt;
        return this;
    }
    /**
     * Sets the initialization code for the forwarder contract.
     *
     * @param {string} address - The address to generate proxy init code for
     * @returns {AddressInitializationBuilder} This transaction builder
     */
    initCode(address) {
        this.validateAddress({ address });
        this.addressInitializationTransaction.initCode = (0, abstract_eth_1.getProxyInitcode)(address);
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.contract, 'Contract address is required');
        (0, assert_1.default)(transaction.baseAddress, 'Base address is required');
        (0, assert_1.default)(transaction.feeAddress, 'Fee address is required');
        (0, assert_1.default)(transaction.salt, 'Salt is required');
        (0, assert_1.default)(transaction.initCode, 'Init code is required');
        this.validateAddress({ address: transaction.contract });
        this.validateAddress({ address: transaction.baseAddress });
        this.validateAddress({ address: transaction.feeAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        const transactionData = this.getAddressInitializationData();
        this.transaction.type = this.transactionType;
        this.addressInitializationTransaction.transactionData = transactionData;
        await this.addressInitializationTransaction.build();
        return this.transaction;
    }
    /**
     * Generates the transaction data for address initialization by encoding the createForwarder method call.
     *
     * @private
     * @returns {string} The encoded transaction data as a hex string
     */
    getAddressInitializationData() {
        const saltBuffer = (0, ethereumjs_util_1.setLengthLeft)((0, ethereumjs_util_1.toBuffer)(this.addressInitializationTransaction.salt), 32);
        const { createForwarderParams, createForwarderTypes } = (0, abstract_eth_1.getCreateForwarderParamsAndTypes)(this.addressInitializationTransaction.baseAddress, saltBuffer, this.addressInitializationTransaction.feeAddress);
        const method = ethereumjs_abi_1.default.methodID('createForwarder', createForwarderTypes);
        const args = ethereumjs_abi_1.default.rawEncode(createForwarderTypes, createForwarderParams);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
}
exports.AddressInitializationBuilder = AddressInitializationBuilder;
