// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/burnNftBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BurnNftBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const burnNftTransaction_1 = require("../transaction/burnNftTransaction");
const utils_1 = __importDefault(require("../utils"));
const constants_1 = require("../constants");
class BurnNftBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new BurnNftBuilder instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new burnNftTransaction_1.BurnNftTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing BurnNftTransaction.
     *
     * @param {BurnNftTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the burn NFT transaction instance.
     *
     * @returns {BurnNftTransaction} The burn NFT transaction
     */
    get burnNftTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for burning NFT.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingWithdraw;
    }
    /**
     * Validates the transaction clauses for burning NFT.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For burn NFT transactions, value must be exactly '0x0'
            if (clause.value !== 0) {
                return false;
            }
            // Check if the data starts with the burn method ID
            if (!clause.data.startsWith(constants_1.BURN_NFT_METHOD_ID)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the token ID for this burn NFT transaction.
     *
     * @param {string} tokenId - The ID of the NFT token to burn
     * @returns {BurnNftBuilder} This transaction builder
     */
    tokenId(tokenId) {
        this.burnNftTransaction.tokenId = tokenId;
        return this;
    }
    /**
     * Sets the staking contract address for this staking tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The staking contract address (required)
     * @returns {StakingBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.burnNftTransaction.stakingContractAddress = address;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        (0, assert_1.default)(transaction.tokenId, 'Token ID is required');
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.burnNftTransaction.build();
        return this.transaction;
    }
}
exports.BurnNftBuilder = BurnNftBuilder;
