// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/claimRewardsBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimRewardsBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const assert_1 = __importDefault(require("assert"));
const transactionBuilder_1 = require("./transactionBuilder");
const claimRewards_1 = require("../transaction/claimRewards");
const constants_1 = require("../constants");
const utils_1 = __importDefault(require("../utils"));
class ClaimRewardsBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new ClaimRewardsBuilder instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new claimRewards_1.ClaimRewardsTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing ClaimRewardsTransaction.
     *
     * @param {ClaimRewardsTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the claim rewards transaction instance.
     *
     * @returns {ClaimRewardsTransaction} The claim rewards transaction
     */
    get claimRewardsTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for claim rewards.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingClaim;
    }
    /**
     * Sets the staking contract address for this claim tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The staking contract address (required)
     * @returns {ClaimRewardsBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.claimRewardsTransaction.stakingContractAddress = address;
        return this;
    }
    /**
     * Sets the token ID for this claim tx.
     *
     * @param {number} levelId - The NFT token ID
     * @returns {DelegateTxnBuilder} This transaction builder
     */
    tokenId(tokenId) {
        this.claimRewardsTransaction.tokenId = tokenId;
        return this;
    }
    /**
     * Validates the transaction clauses for claim rewards transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // Ensure value is '0x0', '0', or 0
            if (!['0x0', '0', 0].includes(clause.value)) {
                return false;
            }
            if (clause.data && clause.data.startsWith(constants_1.CLAIM_STAKING_REWARDS_METHOD_ID)) {
                return true;
            }
            return false;
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.tokenId, 'Token ID is required');
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.claimRewardsTransaction.build();
        return this.transaction;
    }
}
exports.ClaimRewardsBuilder = ClaimRewardsBuilder;
