// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/delegateTxnBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegateTxnBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const delegateClauseTransaction_1 = require("../transaction/delegateClauseTransaction");
const utils_1 = __importDefault(require("../utils"));
class DelegateTxnBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new Delegate Clause txn instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new delegateClauseTransaction_1.DelegateClauseTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing Delegate txn.
     *
     * @param {DelegateClauseTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the staking transaction instance.
     *
     * @returns {DelegateClauseTransaction} The delegate transaction
     */
    get delegateTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for delegate.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingDelegate;
    }
    /**
     * Validates the transaction clauses for delegate transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the staking contract address for this delegate tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The staking contract address (required)
     * @returns {DelegateTxnBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.delegateTransaction.stakingContractAddress = address;
        return this;
    }
    /**
     * Sets the token ID for this delegate tx.
     *
     * @param {number} levelId - The NFT token ID
     * @returns {DelegateTxnBuilder} This transaction builder
     */
    tokenId(tokenId) {
        this.delegateTransaction.tokenId = tokenId;
        return this;
    }
    /**
     * Sets the validator address for this delegate tx.
     * @param {string} address - The validator address
     * @returns {DelegateTxnBuilder} This transaction builder
     */
    validator(address) {
        if (!address) {
            throw new Error('Validator address is required');
        }
        this.validateAddress({ address });
        this.delegateTransaction.validator = address;
        return this;
    }
    /**
     * Sets the transaction data for this delegate tx.
     *
     * @param {string} data - The transaction data
     * @returns {DelegateTxnBuilder} This transaction builder
     */
    transactionData(data) {
        this.delegateTransaction.transactionData = data;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        (0, assert_1.default)(transaction.tokenId, 'Token ID is required');
        (0, assert_1.default)(transaction.validator, 'Validator address is required');
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.delegateTransaction.build();
        return this.transaction;
    }
}
exports.DelegateTxnBuilder = DelegateTxnBuilder;
