// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/exitDelegationBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExitDelegationBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const exitDelegation_1 = require("../transaction/exitDelegation");
const utils_1 = __importDefault(require("../utils"));
const constants_1 = require("../constants");
class ExitDelegationBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new ExitDelegationBuilder instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new exitDelegation_1.ExitDelegationTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing ExitDelegationTransaction.
     *
     * @param {ExitDelegationTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the exit delegation transaction instance.
     *
     * @returns {ExitDelegationTransaction} The exit delegation transaction
     */
    get exitDelegationTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for unstaking.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingUnlock;
    }
    /**
     * Validates the transaction clauses for exit delegation.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For unstaking transactions, value must be exactly '0x0'
            if (clause.value !== 0) {
                return false;
            }
            // Check if the data starts with the exitDelegation method ID
            if (!clause.data.startsWith(constants_1.EXIT_DELEGATION_METHOD_ID)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the token ID for this unstaking transaction.
     *
     * @param {string} tokenId - The ID of the NFT token to unstake
     * @returns {ExitDelegationBuilder} This transaction builder
     */
    tokenId(tokenId) {
        this.exitDelegationTransaction.tokenId = tokenId;
        return this;
    }
    /**
     * Sets the staking contract address for this staking tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The staking contract address (required)
     * @returns {StakingBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.exitDelegationTransaction.stakingContractAddress = address;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        (0, assert_1.default)(transaction.tokenId, 'Token ID is required');
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.exitDelegationTransaction.build();
        return this.transaction;
    }
}
exports.ExitDelegationBuilder = ExitDelegationBuilder;
