// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/flushTokenTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlushTokenTransactionBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("../utils"));
class FlushTokenTransactionBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new FlushTokenTransactionBuilder instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * Initializes the builder with an existing FlushTokenTransaction.
     *
     * @param {FlushTokenTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the flush token transaction instance.
     *
     * @returns {FlushTokenTransaction} The flush token transaction
     */
    get flushTokenTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for flush token.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.FlushTokens;
    }
    /**
     * Validates the transaction clauses for flush token transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For address init transactions, value must be exactly 0
            if (clause.value !== 0) {
                return false;
            }
            const { tokenAddress } = (0, abstract_eth_1.decodeFlushTokensData)(clause.data, clause.to);
            if (!utils_1.default.isValidAddress(tokenAddress)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the token address for this token flush tx.
     *
     * @param {string} address - The token address to be set for the token flush transaction
     * @returns {FlushTokenTransactionBuilder} This transaction builder
     */
    tokenAddress(address) {
        this.validateAddress({ address });
        this.flushTokenTransaction.tokenAddress = address;
        return this;
    }
    /**
     * Sets the forwarder version for this token flush transaction.
     * The forwarder version must be 4 or higher.
     *
     * @param {number} version - The forwarder version to use (must be >= 4)
     * @returns {FlushTokenTransactionBuilder} This transaction builder
     * @throws {BuildTransactionError} When version is less than 4
     */
    forwarderVersion(version) {
        if (version < 4) {
            throw new sdk_core_1.BuildTransactionError(`Invalid forwarder version: ${version}`);
        }
        this.flushTokenTransaction.forwarderVersion = version;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.contract, 'Contract address is required');
        (0, assert_1.default)(transaction.tokenAddress, 'Token address is required');
        this.validateAddress({ address: transaction.contract });
        this.validateAddress({ address: transaction.tokenAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        const transactionData = this.getFlushTokenTransactionData();
        this.transaction.type = this.transactionType;
        this.flushTokenTransaction.transactionData = transactionData;
        await this.flushTokenTransaction.build();
        return this.transaction;
    }
    /**
     * Generates the transaction data for flush token transaction
     *
     * @private
     * @returns {string} The encoded transaction data as a hex string
     */
    getFlushTokenTransactionData() {
        const flushTokenData = (0, abstract_eth_1.flushTokensData)(this.flushTokenTransaction.contract, this.flushTokenTransaction.tokenAddress, this.flushTokenTransaction.forwarderVersion);
        return flushTokenData;
    }
}
exports.FlushTokenTransactionBuilder = FlushTokenTransactionBuilder;
