// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/increaseStakeBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncreaseStakeBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const transactionBuilder_1 = require("./transactionBuilder");
const increaseStakeTransaction_1 = require("../transaction/increaseStakeTransaction");
const utils_1 = __importDefault(require("../utils"));
class IncreaseStakeBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new increase stake txn instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new increaseStakeTransaction_1.IncreaseStakeTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing increase stake txn.
     *
     * @param {IncreaseStakeTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the increase stake transaction instance.
     *
     * @returns {IncreaseStakeTransaction} The increase stake transaction
     */
    get increaseStakeTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for increase stake.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingAdd;
    }
    /**
     * Validates the transaction clauses for increase stake transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the staking contract address for this increase stake tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The contract address (required)
     * @returns {IncreaseStakeBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.increaseStakeTransaction.stakingContractAddress = address;
        return this;
    }
    /**
     * Sets the amount to stake for this increase stake tx (VET amount being sent).
     *
     * @param {string} amount - The amount to stake in wei
     * @returns {IncreaseStakeBuilder} This transaction builder
     */
    amountToStake(amount) {
        this.increaseStakeTransaction.amountToStake = amount;
        return this;
    }
    /**
     * Sets the validator address for this increase stake tx.
     * @param {string} address - The validator address
     * @returns {IncreaseStakeBuilder} This transaction builder
     */
    validator(address) {
        if (!address) {
            throw new Error('Validator address is required');
        }
        this.validateAddress({ address });
        this.increaseStakeTransaction.validator = address;
        return this;
    }
    /**
     * Sets the transaction data for this increase stake tx.
     *
     * @param {string} data - The transaction data
     * @returns {IncreaseStakeBuilder} This transaction builder
     */
    transactionData(data) {
        this.increaseStakeTransaction.transactionData = data;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        (0, assert_1.default)(transaction.validator, 'Validator address is required');
        (0, assert_1.default)(transaction.amountToStake, 'Staking amount is required');
        // Validate staking amount is greater than 0
        const amount = new bignumber_js_1.default(transaction.amountToStake);
        if (amount.isLessThanOrEqualTo(0)) {
            throw new Error('Staking amount must be greater than 0');
        }
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.increaseStakeTransaction.build();
        return this.transaction;
    }
}
exports.IncreaseStakeBuilder = IncreaseStakeBuilder;
