// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/nftTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NFTTransactionBuilder = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("../utils"));
class NFTTransactionBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        super.initBuilder(tx);
    }
    get nftTransaction() {
        return this._transaction;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.SendNFT;
    }
    validateRecipientValue(value) {
        if (value.isNaN()) {
            throw new sdk_core_1.BuildTransactionError('Invalid amount format');
        }
        else if (!value.isEqualTo(1)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be anything other than 1 for NFT transfer');
        }
    }
    recipients(recipients) {
        for (const recipient of recipients) {
            this.validateAddress({ address: recipient.address });
            this.validateRecipientValue(new bignumber_js_1.default(recipient.amount));
        }
        this.transaction.recipients = recipients;
        return this;
    }
    /**
     * Validates the transaction clauses for NFT transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For NFT transactions, the value should be 0
            if (clause.value !== 0) {
                return false;
            }
            const { recipients, sender } = utils_1.default.decodeTransferNFTData(clause.data);
            const recipientAddress = recipients[0].address.toLowerCase();
            if (!recipientAddress || !utils_1.default.isValidAddress(recipientAddress)) {
                return false;
            }
            if (!sender || !utils_1.default.isValidAddress(sender)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    nftCollectionId(nftCollectionId) {
        // nftCollectionId is basically a contract address, so we can use the same validation
        try {
            this.validateAddress({ address: nftCollectionId });
        }
        catch (e) {
            throw new sdk_core_1.BuildTransactionError('Invalid nftCollectionId, must be a valid contract address');
        }
        this.nftTransaction.nftCollectionId = nftCollectionId;
        return this;
    }
    tokenId(tokenId) {
        const tokenIdBN = new bignumber_js_1.default(tokenId);
        if (!tokenIdBN.isInteger() || tokenIdBN.isNegative()) {
            throw new Error('Invalid tokenId, must be a non-negative integer');
        }
        this.nftTransaction.tokenId = tokenIdBN.toFixed(0);
        return this;
    }
}
exports.NFTTransactionBuilder = NFTTransactionBuilder;
