// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/stakingBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakingBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const stakingTransaction_1 = require("../transaction/stakingTransaction");
const utils_1 = __importDefault(require("../utils"));
class StakingBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new StakingBuilder instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new stakingTransaction_1.StakingTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing StakingTransaction.
     *
     * @param {StakingTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the staking transaction instance.
     *
     * @returns {StakingTransaction} The staking transaction
     */
    get stakingTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for staking.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.ContractCall;
    }
    /**
     * Validates the transaction clauses for staking transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For staking transactions, value must be greater than 0
            if (!clause.value || clause.value === '0x0' || clause.value === '0') {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the staking contract address for this staking tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The staking contract address (required)
     * @returns {StakingBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.stakingTransaction.stakingContractAddress = address;
        return this;
    }
    /**
     * Sets the level ID for this staking tx.
     *
     * @param {number} levelId - The level ID for staking
     * @returns {StakingBuilder} This transaction builder
     */
    levelId(levelId) {
        this.stakingTransaction.levelId = levelId;
        return this;
    }
    /**
     * Sets the amount to stake for this staking tx (VET amount being sent).
     *
     * @param {string} amount - The amount to stake in wei
     * @returns {StakingBuilder} This transaction builder
     */
    amountToStake(amount) {
        this.stakingTransaction.amountToStake = amount;
        return this;
    }
    /**
     * Sets the transaction data for this staking tx.
     *
     * @param {string} data - The transaction data
     * @returns {StakingBuilder} This transaction builder
     */
    transactionData(data) {
        this.stakingTransaction.transactionData = data;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        (0, assert_1.default)(transaction.amountToStake, 'Amount to stake is required');
        // Validate amount is a valid number string
        if (transaction.amountToStake) {
            try {
                const bn = new (require('bignumber.js'))(transaction.amountToStake);
                if (!bn.isFinite() || bn.isNaN()) {
                    throw new Error('Invalid character');
                }
            }
            catch (e) {
                throw new Error('Invalid character');
            }
        }
        (0, assert_1.default)(transaction.levelId, 'Level ID is required');
        (0, assert_1.default)(transaction.autorenew, 'Autorenew flag is required');
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.stakingTransaction.build();
        return this.transaction;
    }
}
exports.StakingBuilder = StakingBuilder;
