// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/tokenTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransactionBuilder = void 0;
const ethereumjs_util_1 = require("ethereumjs-util");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("../utils"));
class TokenTransactionBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        this._transaction = tx;
    }
    get tokenTransaction() {
        return this._transaction;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    /**
     * Validates the transaction clauses for flush token transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            // For token transactions, the value should be 0
            if (clause.value !== 0) {
                return false;
            }
            const { address } = utils_1.default.decodeTransferTokenData(clause.data);
            const recipientAddress = (0, ethereumjs_util_1.addHexPrefix)(address.toString()).toLowerCase();
            if (!recipientAddress || !utils_1.default.isValidAddress(recipientAddress)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    tokenAddress(address) {
        this.validateAddress({ address });
        this.tokenTransaction.tokenAddress = address;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        if (!transaction.tokenAddress) {
            throw new Error('Token address is required');
        }
        this.validateAddress({ address: transaction.tokenAddress });
    }
}
exports.TokenTransactionBuilder = TokenTransactionBuilder;
