// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    getNonce() {
        return this.transaction.nonce;
    }
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    /** @inheritdoc */
    set transaction(transaction) {
        this._transaction = transaction;
    }
    initBuilder(tx) {
        this._transaction = tx;
    }
    chainTag(tag) {
        this._transaction.chainTag = tag;
        return this;
    }
    isRecovery(isRecovery) {
        this._transaction.isRecovery = isRecovery;
        return this;
    }
    /**
     * Sets the sender of this transaction.
     *
     * @param {string} senderAddress the account that is sending this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    sender(senderAddress) {
        this.validateAddress({ address: senderAddress });
        this.transaction.sender = senderAddress;
        return this;
    }
    recipients(recipients) {
        for (const recipient of recipients) {
            this.validateAddress({ address: recipient.address });
            this.validateValue(new bignumber_js_1.default(recipient.amount));
        }
        this.transaction.recipients = recipients;
        return this;
    }
    gas(g) {
        this.validateValue(new bignumber_js_1.default(g));
        this.transaction.gas = g;
        return this;
    }
    nonce(n) {
        this.transaction.nonce = n;
        return this;
    }
    expiration(exp) {
        this.transaction.expiration = exp;
        return this;
    }
    dependsOn(dep) {
        this.transaction.dependsOn = dep;
        return this;
    }
    blockRef(ref) {
        this.transaction.blockRef = ref;
        return this;
    }
    gasPriceCoef(coef) {
        this.transaction.gasPriceCoef = coef;
        return this;
    }
    contract(address) {
        this.validateAddress({ address });
        this.transaction.contract = address;
        return this;
    }
    /** @inheritDoc */
    addSenderSignature(signature) {
        this.transaction.addSenderSignature(signature);
    }
    addFeePayerSignature(signature) {
        this.transaction.addFeePayerSignature(signature);
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.transaction.build();
        return this.transaction;
    }
    /** @inheritdoc */
    validateAddress(address, addressFormat) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address ' + address.address);
        }
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        for (const recipient of transaction.recipients) {
            this.validateAddress({ address: recipient.address });
            this.validateValue(new bignumber_js_1.default(recipient.amount));
        }
    }
    isValidRawTransaction(rawTransaction) {
        try {
            const tx = utils_1.default.deserializeTransaction(rawTransaction);
            return this.isValidTransactionClauses(tx.body.clauses);
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    validateRawTransaction(rawTransaction) {
        if (!rawTransaction) {
            throw new sdk_core_1.ParseTransactionError('Invalid raw transaction: Undefined');
        }
        if (!this.isValidRawTransaction(rawTransaction)) {
            throw new sdk_core_1.ParseTransactionError('Invalid raw transaction');
        }
    }
    validateValue(value) {
        if (value.isNaN()) {
            throw new sdk_core_1.BuildTransactionError('Invalid amount format');
        }
        else if (value.isLessThan(0)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be less than zero');
        }
    }
    addFeePayerAddress(address) {
        this.transaction.feePayerAddress = address;
    }
    /** @inheritdoc */
    signImplementation(key) {
        throw new Error('Method not implemented');
    }
    /** @inheritdoc */
    validateKey(key) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        this.validateRawTransaction(rawTransaction);
        this.transaction.fromRawTransaction(rawTransaction);
        return this.transaction;
    }
}
exports.TransactionBuilder = TransactionBuilder;
