// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilder/validatorRegistrationBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorRegistrationBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const transactionBuilder_1 = require("./transactionBuilder");
const validatorRegistrationTransaction_1 = require("../transaction/validatorRegistrationTransaction");
const utils_1 = __importDefault(require("../utils"));
class ValidatorRegistrationBuilder extends transactionBuilder_1.TransactionBuilder {
    /**
     * Creates a new add validation Clause txn instance.
     *
     * @param {Readonly<CoinConfig>} _coinConfig - The coin configuration object
     */
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new validatorRegistrationTransaction_1.ValidatorRegistrationTransaction(_coinConfig);
    }
    /**
     * Initializes the builder with an existing validation registration txn.
     *
     * @param {ValidatorRegistrationTransaction} tx - The transaction to initialize the builder with
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /**
     * Gets the staking transaction instance.
     *
     * @returns {ValidatorRegistrationTransaction} The validator registration transaction
     */
    get validatorRegistrationTransaction() {
        return this._transaction;
    }
    /**
     * Gets the transaction type for validator registration.
     *
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingLock;
    }
    /**
     * Validates the transaction clauses for validator registration transaction.
     * @param {TransactionClause[]} clauses - The transaction clauses to validate.
     * @returns {boolean} - Returns true if the clauses are valid, false otherwise.
     */
    isValidTransactionClauses(clauses) {
        try {
            if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
                return false;
            }
            const clause = clauses[0];
            if (!clause.to || !utils_1.default.isValidAddress(clause.to)) {
                return false;
            }
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Sets the staking contract address for this validator registration tx.
     * The address must be explicitly provided to ensure the correct contract is used.
     *
     * @param {string} address - The contract address (required)
     * @returns {ValidatorRegistrationBuilder} This transaction builder
     * @throws {Error} If no address is provided
     */
    stakingContractAddress(address) {
        if (!address) {
            throw new Error('Staking contract address is required');
        }
        this.validateAddress({ address });
        this.validatorRegistrationTransaction.stakingContractAddress = address;
        return this;
    }
    /**
     * Sets the staking period for this validator registration tx.
     *
     * @param {number} period - The staking period
     * @returns {ValidatorRegistrationBuilder} This transaction builder
     */
    stakingPeriod(period) {
        this.validatorRegistrationTransaction.stakingPeriod = period;
        return this;
    }
    /**
     * Sets the amount to stake for this validator registration tx (VET amount being sent).
     *
     * @param {string} amount - The amount to stake in wei
     * @returns {ValidatorRegistrationBuilder} This transaction builder
     */
    amountToStake(amount) {
        this.validatorRegistrationTransaction.amountToStake = amount;
        return this;
    }
    /**
     * Sets the validator address for this validator registration tx.
     * @param {string} address - The validator address
     * @returns {ValidatorRegistrationBuilder} This transaction builder
     */
    validator(address) {
        if (!address) {
            throw new Error('Validator address is required');
        }
        this.validateAddress({ address });
        this.validatorRegistrationTransaction.validator = address;
        return this;
    }
    /**
     * Sets the transaction data for this validator registration tx.
     *
     * @param {string} data - The transaction data
     * @returns {ValidatorRegistrationBuilder} This transaction builder
     */
    transactionData(data) {
        this.validatorRegistrationTransaction.transactionData = data;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        (0, assert_1.default)(transaction.stakingContractAddress, 'Staking contract address is required');
        (0, assert_1.default)(transaction.stakingPeriod, 'Staking period is required');
        (0, assert_1.default)(transaction.validator, 'Validator address is required');
        (0, assert_1.default)(transaction.amountToStake, 'Staking amount is required');
        // Validate staking amount is within allowed range
        const amountInVET = new bignumber_js_1.default(transaction.amountToStake).dividedBy(new bignumber_js_1.default(10).pow(18));
        if (amountInVET.isLessThan(25000000) || amountInVET.isGreaterThan(600000000)) {
            throw new Error('Staking amount must be between 25M and 600M VET');
        }
        this.validateAddress({ address: transaction.stakingContractAddress });
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.type = this.transactionType;
        await this.validatorRegistrationTransaction.build();
        return this.transaction;
    }
}
exports.ValidatorRegistrationBuilder = ValidatorRegistrationBuilder;
