// @bitgo/utxo-core/dist/cjs/test/bip322/bip322.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BIP322_FIXTURE_HELLO_WORLD_TOSIGN_PSBT = exports.BIP322_FIXTURE_HELLO_WORLD_TOSPEND_TX = exports.BIP322_PAYMENT_P2WPKH_FIXTURE = exports.BIP322_PRV_FIXTURE = exports.BIP322_WIF_FIXTURE = void 0;
const utxo_lib_1 = require("@bitgo/utxo-lib");
const bip322_1 = require("../../src/bip322");
exports.BIP322_WIF_FIXTURE = 'L3VFeEujGtevx9w18HD1fhRbCH67Az2dpCymeRE1SoPK6XQtaN2k';
exports.BIP322_PRV_FIXTURE = utxo_lib_1.ECPair.fromWIF(exports.BIP322_WIF_FIXTURE);
exports.BIP322_PAYMENT_P2WPKH_FIXTURE = utxo_lib_1.payments.p2wpkh({
    address: 'bc1q9vza2e8x573nczrlzms0wvx3gsqjx7vavgkx0l',
});
exports.BIP322_FIXTURE_HELLO_WORLD_TOSPEND_TX = (0, bip322_1.buildToSpendTransaction)(exports.BIP322_PAYMENT_P2WPKH_FIXTURE.output, Buffer.from('Hello World'));
exports.BIP322_FIXTURE_HELLO_WORLD_TOSIGN_PSBT = (0, bip322_1.createBaseToSignPsbt)();
(0, bip322_1.addBip322Input)(exports.BIP322_FIXTURE_HELLO_WORLD_TOSIGN_PSBT, 'Hello World', {
    scriptPubKey: exports.BIP322_PAYMENT_P2WPKH_FIXTURE.output,
});
