// @bitgo/utxo-core/dist/cjs/test/bip322/toSpend.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const bip322_1 = require("../../src/bip322");
const bip322_utils_1 = require("./bip322.utils");
describe('to_spend', function () {
    describe('Message hashing', function () {
        // Test vectors from BIP322
        // Source: https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki#message-hashing
        const fixtures = [
            {
                message: '',
                hash: 'c90c269c4f8fcbe6880f72a721ddfbf1914268a794cbb21cfafee13770ae19f1',
            },
            {
                message: 'Hello World',
                hash: 'f0eb03b1a75ac6d9847f55c624a99169b5dccba2a31f5b23bea77ba270de0a7a',
            },
        ];
        fixtures.forEach(({ message, hash }) => {
            it(`should hash the message "${message}"`, function () {
                const result = (0, bip322_1.hashMessageWithTag)(Buffer.from(message));
                assert_1.default.deepStrictEqual(result.toString('hex'), hash, `Hash for message "${message}" does not match expected value`);
            });
        });
    });
    describe('build to_spend transaction', function () {
        const scriptPubKey = bip322_utils_1.BIP322_PAYMENT_P2WPKH_FIXTURE.output;
        // Source: https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki#transaction-hashes
        const fixtures = [
            {
                message: '',
                txid: 'c5680aa69bb8d860bf82d4e9cd3504b55dde018de765a91bb566283c545a99a7',
            },
            {
                message: 'Hello World',
                txid: 'b79d196740ad5217771c1098fc4a4b51e0535c32236c71f1ea4d61a2d603352b',
            },
        ];
        fixtures.forEach(({ message, txid }) => {
            it(`should build a to_spend transaction for message "${message}"`, function () {
                const result = (0, bip322_1.buildToSpendTransaction)(scriptPubKey, Buffer.from(message));
                const computedTxid = result.getId();
                assert_1.default.strictEqual(computedTxid, txid, `Transaction ID for message "${message}" does not match expected value`);
            });
        });
    });
    describe('buildToSpendTransactionFromChainAndIndex', function () {
        describe('should build a to_spend transaction for a non-Taproot chain', function () {
            function run(chain) {
                it(`scriptType: ${utxo_lib_1.bitgo.scriptTypeForChain(chain)}, chain ${chain}`, function () {
                    const tx = (0, bip322_1.buildToSpendTransactionFromChainAndIndex)(utxo_lib_1.testutil.getDefaultWalletKeys(), 20, 0, Buffer.from('Hello World'));
                    const expectedScriptPubKey = utxo_lib_1.bitgo.outputScripts
                        .createOutputScript2of3(utxo_lib_1.testutil.getDefaultWalletKeys().deriveForChainAndIndex(20, 0).publicKeys, 'p2wsh')
                        .scriptPubKey.toString();
                    const scriptPubKeyFromTx = tx.outs[0].script.toString();
                    assert_1.default.deepStrictEqual(scriptPubKeyFromTx, expectedScriptPubKey, 'ScriptPubKey does not match expected value');
                });
            }
            [0, 1, 10, 11, 20, 21].forEach((chain) => {
                run(chain);
            });
        });
    });
});
