// @bitgo/sdk-coin-vet/dist/src/vet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vet = void 0;
const _ = __importStar(require("lodash"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const blake2b_1 = __importDefault(require("@bitgo/blake2b"));
const assert_1 = __importDefault(require("assert"));
const axios_1 = __importDefault(require("axios"));
const logger_1 = require("@bitgo/logger");
const sdk_core_1 = require("@vechain/sdk-core");
const sdk_core_2 = require("@bitgo/sdk-core");
const mpc = __importStar(require("@bitgo/sdk-lib-mpc"));
const statics_1 = require("@bitgo/statics");
const utils_1 = __importDefault(require("./lib/utils"));
const secp256k1_1 = require("@bitgo/secp256k1");
const crypto_1 = require("crypto");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
const constants_1 = require("./lib/constants");
/**
 * Full Name: Vechain
 * Docs: https://docs.vechain.org/
 * GitHub : https://github.com/vechain
 */
class Vet extends sdk_core_2.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Vet(bitgo, staticsCoin);
    }
    /**
     * Factor between the coin's base unit and its smallest sub division
     */
    getBaseFactor() {
        return 1e18;
    }
    getChain() {
        return 'vet';
    }
    getFamily() {
        return 'vet';
    }
    getFullName() {
        return 'VeChain';
    }
    valuelessTransferAllowed() {
        return true;
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_2.multisigTypes.tss;
    }
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    allowsAccountConsolidations() {
        return true;
    }
    async verifyTransaction(params) {
        const { txPrebuild: txPrebuild, txParams: txParams } = params;
        const txHex = txPrebuild.txHex;
        if (!txHex) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const explainedTx = await this.explainTransaction({ txHex });
        if (!explainedTx) {
            throw new Error('failed to explain transaction');
        }
        if (txParams.recipients !== undefined && txParams.recipients.length > 0) {
            const filteredRecipients = txParams.recipients?.map((recipient) => {
                return {
                    address: recipient.address.toLowerCase(),
                    amount: BigInt(recipient.amount),
                };
            });
            const filteredOutputs = explainedTx.outputs.map((output) => {
                return {
                    address: output.address.toLowerCase(),
                    amount: BigInt(output.amount),
                };
            });
            if (!_.isEqual(filteredOutputs, filteredRecipients)) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            let totalAmount = new bignumber_js_1.default(0);
            for (const recipients of txParams.recipients) {
                totalAmount = totalAmount.plus(recipients.amount);
            }
            if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                throw new Error('Tx total amount does not match with expected total amount field');
            }
        }
        return true;
    }
    /**
     * Verify that an address belongs to this wallet.
     *
     * @param {TssVerifyVetAddressOptions} params - Verification parameters
     * @returns {Promise<boolean>} True if address belongs to wallet
     * @throws {InvalidAddressError} If address format is invalid
     * @throws {Error} If invalid wallet version or missing parameters
     */
    async isWalletAddress(params) {
        const { address, baseAddress, walletVersion } = params;
        if (address && !this.isValidAddress(address)) {
            throw new sdk_core_2.InvalidAddressError(`invalid address: ${address}`);
        }
        if (walletVersion !== 6) {
            throw new Error(`VET only supports wallet version 6, but got version ${walletVersion}`);
        }
        const isVerifyingBaseAddress = baseAddress && address.toLowerCase() === baseAddress.toLowerCase();
        if (isVerifyingBaseAddress) {
            const index = typeof params.index === 'string' ? parseInt(params.index, 10) : params.index;
            if (index !== 0) {
                throw new Error(`Base address verification requires index 0, but got index ${params.index}.`);
            }
        }
        const result = await (0, sdk_core_2.verifyMPCWalletAddress)({ ...params, keyCurve: 'secp256k1' }, this.isValidAddress.bind(this), (pubKey) => {
            const keyPair = new abstract_eth_1.KeyPair({ pub: pubKey });
            return keyPair.getAddress();
        });
        if (!result) {
            throw new sdk_core_2.InvalidAddressError(`invalid address: ${address}`);
        }
        return true;
    }
    async parseTransaction(params) {
        const transactionExplanation = await this.explainTransaction({ txHex: params.txHex });
        if (!transactionExplanation) {
            throw new Error('Invalid transaction');
        }
        return {
            inputs: [
                {
                    address: transactionExplanation.sender,
                    amount: transactionExplanation.outputAmount,
                },
            ],
            outputs: [
                {
                    address: transactionExplanation.outputs[0].address,
                    amount: transactionExplanation.outputs[0].amount,
                },
            ],
        };
    }
    /**
     * Explain a Vechain transaction
     * @param params
     */
    async explainTransaction(params) {
        let rebuiltTransaction;
        try {
            rebuiltTransaction = await this.rebuildTransaction(params.txHex);
        }
        catch {
            return undefined;
        }
        return rebuiltTransaction.explainTransaction();
    }
    generateKeyPair(seed) {
        if (!seed) {
            // An extended private key has both a normal 256 bit private key and a 256
            // bit chain code, both of which must be random. 512 bits is therefore the
            // maximum entropy and gives us maximum security against cracking.
            seed = (0, crypto_1.randomBytes)(512 / 8);
        }
        const extendedKey = secp256k1_1.bip32.fromSeed(seed);
        const xpub = extendedKey.neutered().toBase58();
        return {
            pub: xpub,
            prv: extendedKey.toBase58(),
        };
    }
    isValidPub(pub) {
        let valid = true;
        try {
            new abstract_eth_1.KeyPair({ pub });
        }
        catch (e) {
            valid = false;
        }
        return valid;
    }
    isValidAddress(address) {
        return utils_1.default.isValidAddress(address);
    }
    signTransaction(params) {
        throw new Error('Method not implemented.');
    }
    getTxBuilderFactory() {
        return new lib_1.TransactionBuilderFactory(this._staticsCoin);
    }
    async rebuildTransaction(txHex) {
        const txBuilderFactory = this.getTxBuilderFactory();
        try {
            const txBuilder = txBuilderFactory.from(txHex);
            return txBuilder.transaction;
        }
        catch {
            throw new Error('Failed to rebuild transaction');
        }
    }
    /** @inheritDoc */
    auditDecryptedKey(params) {
        /** https://bitgoinc.atlassian.net/browse/COIN-4213 */
        throw new Error('Method not implemented.');
    }
    /**
     * Function to get coin specific hash function used to generate transaction digests.
     * @returns {@see Hash} hash function if implemented, otherwise throws exception
     */
    getHashFunction() {
        const blake = (0, blake2b_1.default)(32);
        // We return an object that mimics the Hash interface
        return {
            update(data) {
                blake.update(data);
                return this;
            },
            digest() {
                const uint8Result = blake.digest();
                return Buffer.from(uint8Result);
            },
        };
    }
    buildNftTransferData(params) {
        const { recipientAddress, fromAddress, tokenContractAddress } = params;
        if (!utils_1.default.isValidAddress(recipientAddress)) {
            throw new sdk_core_2.InvalidAddressError('Invalid recipient address');
        }
        if (!utils_1.default.isValidAddress(fromAddress)) {
            throw new sdk_core_2.InvalidAddressError('Invalid from address');
        }
        if (!utils_1.default.isValidAddress(tokenContractAddress)) {
            throw new sdk_core_2.InvalidAddressError('Invalid NFT contract address address');
        }
        switch (params.type) {
            case 'ERC721': {
                const tokenId = params.tokenId;
                return {
                    tokenType: sdk_core_2.TokenType.ERC721,
                    tokenQuantity: '1', // This NFT standard will always have quantity of 1
                    tokenContractAddress,
                    tokenId,
                };
            }
            default:
                throw new sdk_core_2.NotImplementedError(`NFT type ${params.type} not supported on ${this.getChain()}`);
        }
    }
    /**
     * Broadcasts a signed transaction to the VeChain network.
     *
     * @param {BaseBroadcastTransactionOptions} payload - The payload containing the serialized signed transaction.
     * @param {string} payload.serializedSignedTransaction - The serialized signed transaction to broadcast.
     * @returns {Promise<BaseBroadcastTransactionResult>} A promise that resolves to an empty object if the broadcast is successful.
     * @throws {Error} If the broadcast fails, an error is thrown with the failure message.
     */
    async broadcastTransaction(payload) {
        const baseUrl = this.getPublicNodeUrl();
        const url = `${baseUrl}/transactions`;
        // The body should be a JSON object with a 'raw' key
        const requestBody = {
            raw: payload.serializedSignedTransaction,
        };
        try {
            await axios_1.default.post(url, requestBody);
            return {};
        }
        catch (error) {
            throw new Error(`Failed to broadcast transaction: ${error.message}`);
        }
    }
    /** @inheritDoc */
    async recover(params) {
        if (params.tokenContractAddress) {
            return this.recoverTokens(params);
        }
        try {
            if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
                throw new Error('invalid recoveryDestination');
            }
            const isUnsignedSweep = !params.userKey && !params.backupKey && !params.walletPassphrase;
            let publicKey;
            let userKeyShare, backupKeyShare, commonKeyChain;
            const MPC = new sdk_core_2.Ecdsa();
            if (isUnsignedSweep) {
                const bitgoKey = params.bitgoKey;
                if (!bitgoKey) {
                    throw new Error('missing bitgoKey');
                }
                const hdTree = new mpc.Secp256k1Bip32HdTree();
                const derivationPath = 'm/0';
                const derivedPub = hdTree.publicDerive({
                    pk: mpc.bigIntFromBufferBE(Buffer.from(bitgoKey.slice(0, 66), 'hex')),
                    chaincode: mpc.bigIntFromBufferBE(Buffer.from(bitgoKey.slice(66), 'hex')),
                }, derivationPath);
                publicKey = mpc.bigIntToBufferBE(derivedPub.pk).toString('hex');
            }
            else {
                if (!params.userKey) {
                    throw new Error('missing userKey');
                }
                if (!params.backupKey) {
                    throw new Error('missing backupKey');
                }
                if (!params.walletPassphrase) {
                    throw new Error('missing wallet passphrase');
                }
                const userKey = params.userKey.replace(/\s/g, '');
                const backupKey = params.backupKey.replace(/\s/g, '');
                ({ userKeyShare, backupKeyShare, commonKeyChain } = await sdk_core_2.ECDSAUtils.getMpcV2RecoveryKeyShares(userKey, backupKey, params.walletPassphrase));
                publicKey = MPC.deriveUnhardened(commonKeyChain, 'm/0').slice(0, 66);
            }
            if (!publicKey) {
                throw new Error('failed to derive public key');
            }
            const backupKeyPair = new abstract_eth_1.KeyPair({ pub: publicKey });
            const baseAddress = backupKeyPair.getAddress();
            const tx = await this.buildRecoveryTransaction({
                baseAddress,
                params,
            });
            const signableHex = await tx.signablePayload;
            const serializedTxHex = await tx.toBroadcastFormat();
            if (isUnsignedSweep) {
                return {
                    txHex: serializedTxHex,
                    coin: this.getChain(),
                };
            }
            const signableMessage = this.getHashFunction().update(signableHex).digest();
            const signatureObj = await sdk_core_2.ECDSAUtils.signRecoveryMpcV2(signableMessage, userKeyShare, backupKeyShare, commonKeyChain);
            const signature = Buffer.from(signatureObj.r + signatureObj.s + (signatureObj.recid === 0 ? '00' : '01'), 'hex');
            const txBuilder = this.getTxBuilderFactory().getTransferBuilder();
            await txBuilder.from(serializedTxHex);
            txBuilder.isRecovery(true);
            await txBuilder.addSenderSignature(signature);
            const signedTx = await txBuilder.build();
            return {
                id: signedTx.id,
                tx: signedTx.toBroadcastFormat(),
            };
        }
        catch (error) {
            throw new Error(`Error during Vechain recovery: ${error.message || error}`);
        }
    }
    /**
     * Returns the public node URL for the VeChain network.
     * @returns {string} The URL of the public VeChain node.
     */
    getPublicNodeUrl() {
        return sdk_core_2.Environments[this.bitgo.getEnv()].vetNodeUrl;
    }
    /**
     * Calculates the transaction fee based on the estimated gas limit and fee estimate data.
     * @param {FeeEstimateData} feeEstimateData - The fee estimate data.
     * @param {BigNumber} estimatedGasLimit - The estimated gas limit for the transaction.
     * @returns {BigNumber} The calculated transaction fee.
     */
    calculateFee(feeEstimateData, estimatedGasLimit) {
        const gasLimit = estimatedGasLimit;
        const adjustmentFactor = new bignumber_js_1.default(1).plus(new bignumber_js_1.default(feeEstimateData.gasPriceCoef)
            .dividedBy(new bignumber_js_1.default(feeEstimateData.coefDivisor))
            .decimalPlaces(18, bignumber_js_1.default.ROUND_DOWN));
        const adjustedGasPrice = new bignumber_js_1.default(feeEstimateData.gasUnitPrice).times(adjustmentFactor);
        return gasLimit.times(adjustedGasPrice).integerValue(bignumber_js_1.default.ROUND_CEIL);
    }
    /**
     * Ensures that the given address has sufficient VTHO balance to cover the transaction fee.
     * @param {string} baseAddress - The address to check for VTHO balance.
     * @param {BigNumber} requiredGasUnits - The required gas units for the transaction.
     * @throws {Error} If the VTHO balance is insufficient or if there's an error checking the balance.
     */
    async ensureVthoBalanceForFee(baseAddress, requiredGasUnits) {
        const vthoTokenAddress = '0x0000000000000000000000000000456E65726779'; // VTHO token contract address
        try {
            const vthoBalance = await this.getBalance(baseAddress, vthoTokenAddress);
            const requiredFee = this.calculateFee(constants_1.feeEstimateData, requiredGasUnits);
            if (vthoBalance.isLessThan(requiredFee)) {
                throw new Error(`Insufficient VTHO balance for fees. Required: ${requiredFee.toString()}, Available: ${vthoBalance.toString()}`);
            }
        }
        catch (error) {
            throw new Error(`Failed to ensure VTHO balance: ${error.message}`);
        }
    }
    /**
     * Fetches the balance for a given Vechain address.
     *
     * @param address The Vechain address (e.g., "0x...") to check.
     * @param tokenContractAddress (Optional) The contract address of a VIP180 token.
     * @returns A Promise that resolves to a BigNumber instance of the balance.
     */
    async getBalance(address, tokenContractAddress) {
        const baseUrl = this.getPublicNodeUrl();
        if (!tokenContractAddress) {
            const url = `${baseUrl}/accounts/${address}`;
            try {
                const response = await axios_1.default.get(url);
                // The 'balance' is returned as a hex string.
                const balance = new bignumber_js_1.default(response.data.balance);
                return balance;
            }
            catch (error) {
                throw new Error('Failed to get native balance.');
            }
        }
        const url = `${baseUrl}/accounts/*`;
        // Construct the ABI-encoded data for the 'balanceOf(address)' call
        // 1. Function selector for 'balanceOf(address)': '0x70a08231'
        // 2. Padded address: The address, stripped of '0x', left-padded with zeros to 64 chars
        const paddedAddress = address.startsWith('0x') ? address.substring(2).padStart(64, '0') : address.padStart(64, '0');
        const data = `0x70a08231${paddedAddress}`;
        const requestBody = {
            clauses: [
                {
                    to: tokenContractAddress, // The token contract address
                    value: '0x0',
                    data: data, // The 'balanceOf' call
                },
            ],
        };
        try {
            const response = await axios_1.default.post(url, requestBody);
            const simResponse = response.data;
            // Validate response and extract the balance data
            if (!simResponse || !Array.isArray(simResponse) || simResponse.length === 0 || !simResponse[0].data) {
                throw new Error('Invalid simulation response from VeChain node');
            }
            // The returned data is the hex-encoded balance
            return new bignumber_js_1.default(simResponse[0].data);
        }
        catch (error) {
            logger_1.logger.error('Error fetching token balance:', error);
            throw new Error(`Failed to get token balance: ${error.message}`);
        }
    }
    /**
     * Retrieves the block reference from the VeChain network.
     * @returns {Promise<string>} A promise that resolves to the block reference string.
     * @throws {Error} If there's an error fetching the block reference or if the response is invalid.
     */
    async getBlockRef() {
        const baseUrl = this.getPublicNodeUrl();
        const url = `${baseUrl}/blocks/best`;
        try {
            const response = await axios_1.default.get(url);
            const data = response.data;
            // Validate the response data
            if (!data || !data.id) {
                throw new Error('Invalid response from the VeChain node');
            }
            // Return the first 18 characters of the block ID
            return data.id.slice(0, 18);
        }
        catch (error) {
            // Rethrow or return a sensible default
            throw new Error('Failed to get block ref: ');
        }
    }
    /**
     * Generates a random nonce for use in transactions.
     * @returns {string} A hexadecimal string representing the random nonce.
     */
    getRandomNonce() {
        return '0x' + (0, crypto_1.randomBytes)(8).toString('hex');
    }
    /**
     * Estimates the gas required for a set of transaction clauses.
     * @param {TransactionClause[]} clauses - An array of transaction clauses.
     * @param {string} caller - The address of the transaction caller.
     * @returns {Promise<BigNumber>} A promise that resolves to the estimated gas amount.
     * @throws {Error} If the clauses are invalid, the caller is not provided, or if there's an error in gas estimation.
     */
    async estimateGas(clauses, caller) {
        if (!clauses || !Array.isArray(clauses) || clauses.length === 0) {
            throw new Error('Clauses must be a non-empty array');
        }
        if (!caller) {
            throw new Error('Caller address is required');
        }
        const baseUrl = this.getPublicNodeUrl();
        const url = `${baseUrl}/accounts/*`;
        const requestBody = {
            clauses: clauses,
            caller: caller,
        };
        try {
            const response = await axios_1.default.post(url, requestBody);
            const simResponse = response.data;
            if (!simResponse || !Array.isArray(simResponse)) {
                throw new Error('Invalid simulation response from VeChain node');
            }
            const totalSimulatedGas = simResponse.reduce((sum, result) => sum + (result.gasUsed || 0), 0);
            const intrinsicGas = Number(sdk_core_1.Transaction.intrinsicGas(clauses).wei);
            const totalGas = Math.ceil(intrinsicGas + (totalSimulatedGas !== 0 ? totalSimulatedGas + 15000 : 0));
            return new bignumber_js_1.default(totalGas);
        }
        catch (error) {
            throw new Error(`Failed to estimate gas: ${error.message}`);
        }
    }
    /**
     * Builds a recovery transaction for the given address.
     * @param {Object} buildParams - The parameters for building the recovery transaction.
     * @param {string} buildParams.baseAddress - The address to recover funds from.
     * @param {RecoverOptions} buildParams.params - The recovery options.
     * @returns {Promise<Transaction>} A promise that resolves to the built recovery transaction.
     * @throws {Error} If there's no VET balance to recover or if there's an error building the transaction.
     */
    async buildRecoveryTransaction(buildParams) {
        const { baseAddress, params } = buildParams;
        const balance = await this.getBalance(baseAddress);
        if (balance.isLessThanOrEqualTo(0)) {
            throw new Error(`no VET balance to recover for address ${baseAddress}`);
        }
        const recipients = [
            {
                address: params.recoveryDestination,
                amount: balance.toString(),
            },
        ];
        const blockRef = await this.getBlockRef();
        const txBuilder = this.getTxBuilderFactory().getTransferBuilder();
        txBuilder.chainTag(this.bitgo.getEnv() === 'prod' ? 0x4a : 0x27);
        txBuilder.recipients(recipients);
        txBuilder.sender(baseAddress);
        txBuilder.addFeePayerAddress(baseAddress);
        txBuilder.gas(Number(constants_1.AVG_GAS_UNITS));
        txBuilder.blockRef(blockRef);
        txBuilder.expiration(constants_1.EXPIRATION);
        txBuilder.gasPriceCoef(Number(constants_1.GAS_PRICE_COEF));
        txBuilder.nonce(this.getRandomNonce());
        txBuilder.isRecovery(true);
        let tx = (await txBuilder.build());
        const clauses = tx.clauses;
        const actualGasUnits = await this.estimateGas(clauses, baseAddress);
        await this.ensureVthoBalanceForFee(baseAddress, actualGasUnits);
        txBuilder.gas(actualGasUnits.toNumber());
        tx = (await txBuilder.build());
        return tx;
    }
    async recoverTokens(params) {
        try {
            if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
                throw new Error('invalid recoveryDestination');
            }
            if (!params.tokenContractAddress || !this.isValidAddress(params.tokenContractAddress)) {
                throw new Error('invalid tokenContractAddress');
            }
            const isUnsignedSweep = !params.userKey && !params.backupKey && !params.walletPassphrase;
            let publicKey;
            let userKeyShare, backupKeyShare, commonKeyChain;
            const MPC = new sdk_core_2.Ecdsa();
            if (isUnsignedSweep) {
                const bitgoKey = params.bitgoKey;
                if (!bitgoKey) {
                    throw new Error('missing bitgoKey');
                }
                const hdTree = new mpc.Secp256k1Bip32HdTree();
                const derivationPath = 'm/0';
                const derivedPub = hdTree.publicDerive({
                    pk: mpc.bigIntFromBufferBE(Buffer.from(bitgoKey.slice(0, 66), 'hex')),
                    chaincode: mpc.bigIntFromBufferBE(Buffer.from(bitgoKey.slice(66), 'hex')),
                }, derivationPath);
                publicKey = mpc.bigIntToBufferBE(derivedPub.pk).toString('hex');
            }
            else {
                if (!params.userKey) {
                    throw new Error('missing userKey');
                }
                if (!params.backupKey) {
                    throw new Error('missing backupKey');
                }
                if (!params.walletPassphrase) {
                    throw new Error('missing wallet passphrase');
                }
                const userKey = params.userKey.replace(/\s/g, '');
                const backupKey = params.backupKey.replace(/\s/g, '');
                ({ userKeyShare, backupKeyShare, commonKeyChain } = await sdk_core_2.ECDSAUtils.getMpcV2RecoveryKeyShares(userKey, backupKey, params.walletPassphrase));
                publicKey = MPC.deriveUnhardened(commonKeyChain, 'm/0').slice(0, 66);
            }
            if (!publicKey) {
                throw new Error('failed to derive public key');
            }
            const backupKeyPair = new abstract_eth_1.KeyPair({ pub: publicKey });
            const baseAddress = backupKeyPair.getAddress();
            const tx = await this.buildTokenRecoveryTransaction({
                baseAddress,
                params,
            });
            const signableHex = await tx.signablePayload;
            const serializedTxHex = await tx.toBroadcastFormat();
            if (isUnsignedSweep) {
                return {
                    txHex: serializedTxHex,
                    coin: this.getChain(),
                };
            }
            const signableMessage = this.getHashFunction().update(signableHex).digest();
            const signatureObj = await sdk_core_2.ECDSAUtils.signRecoveryMpcV2(signableMessage, userKeyShare, backupKeyShare, commonKeyChain);
            const signature = Buffer.from(signatureObj.r + signatureObj.s + (signatureObj.recid === 0 ? '00' : '01'), 'hex');
            const tokenTransaction = new lib_1.TokenTransaction(statics_1.coins.get(this.getChain()));
            const txBuilder = this.getTxBuilderFactory().getTokenTransactionBuilder(tokenTransaction);
            await txBuilder.from(serializedTxHex);
            txBuilder.isRecovery(true);
            await txBuilder.addSenderSignature(signature);
            const signedTx = await txBuilder.build();
            return {
                id: signedTx.id,
                tx: signedTx.toBroadcastFormat(),
            };
        }
        catch (error) {
            throw new Error(`Error during Vechain token recovery: ${error.message || error}`);
        }
    }
    async buildTokenRecoveryTransaction(buildParams) {
        const { baseAddress, params } = buildParams;
        const tokenContractAddress = params.tokenContractAddress;
        (0, assert_1.default)(tokenContractAddress, 'tokenContractAddress is required for token recovery');
        const balance = await this.getBalance(baseAddress, tokenContractAddress);
        //replace with get balance function
        if (balance.isLessThanOrEqualTo(0)) {
            throw new Error(`no token balance to recover for address ${baseAddress} contract address ${tokenContractAddress}`);
        }
        // create the recipients here so that we can build the clauses for gas estimation
        const roughFeeEstimate = this.calculateFee(constants_1.feeEstimateData, new bignumber_js_1.default(51390));
        let recipients = [
            {
                address: params.recoveryDestination,
                amount: balance.minus(roughFeeEstimate).toString(),
            },
        ];
        const blockRef = await this.getBlockRef();
        const tokenTransaction = new lib_1.TokenTransaction(statics_1.coins.get(this.getChain()));
        const txBuilder = this.getTxBuilderFactory().getTokenTransactionBuilder(tokenTransaction);
        txBuilder.tokenAddress(tokenContractAddress);
        txBuilder.chainTag(this.bitgo.getEnv() === 'prod' ? 0x4a : 0x27);
        txBuilder.recipients(recipients);
        txBuilder.sender(baseAddress);
        txBuilder.addFeePayerAddress(baseAddress);
        txBuilder.gas(Number(constants_1.AVG_GAS_UNITS));
        txBuilder.blockRef(blockRef);
        txBuilder.expiration(constants_1.EXPIRATION);
        txBuilder.gasPriceCoef(Number(constants_1.GAS_PRICE_COEF));
        txBuilder.nonce(this.getRandomNonce());
        txBuilder.isRecovery(true);
        let tx = (await txBuilder.build());
        const clauses = tx.clauses;
        const actualGasUnits = await this.estimateGas(clauses, baseAddress);
        await this.ensureVthoBalanceForFee(baseAddress, actualGasUnits);
        const requiredFee = this.calculateFee(constants_1.feeEstimateData, actualGasUnits);
        // create the final recipients with the fee deducted
        recipients = [
            {
                address: params.recoveryDestination,
                amount: balance.minus(requiredFee).toString(),
            },
        ];
        txBuilder.recipients(recipients);
        txBuilder.gas(actualGasUnits.toNumber());
        tx = (await txBuilder.build());
        return tx;
    }
}
exports.Vet = Vet;
