// @bitgo/sdk-coin-vet/dist/src/vetNFTCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VetNFTCollection = void 0;
const vet_1 = require("./vet");
const statics_1 = require("@bitgo/statics");
class VetNFTCollection extends vet_1.Vet {
    constructor(bitgo, nftCollectionConfig) {
        const staticsCoin = nftCollectionConfig.network === 'Mainnet' ? statics_1.coins.get('vet') : statics_1.coins.get('tvet');
        super(bitgo, staticsCoin);
        this.nftCollectionConfig = nftCollectionConfig;
    }
    static createNFTCollectionConstructor(config) {
        return (bitgo) => new VetNFTCollection(bitgo, config);
    }
    static createNFTCollectionConstructors(nftCollectionConfigs = [
        ...statics_1.tokens.bitcoin.vet.nftCollections,
        ...statics_1.tokens.testnet.vet.nftCollections,
    ]) {
        const nftCollectionCtors = [];
        for (const config of nftCollectionConfigs) {
            const nftCollectionConstructor = VetNFTCollection.createNFTCollectionConstructor(config);
            nftCollectionCtors.push({ name: config.type, coinConstructor: nftCollectionConstructor });
        }
        return nftCollectionCtors;
    }
    get name() {
        return this.nftCollectionConfig.name;
    }
    get coin() {
        return this.nftCollectionConfig.coin;
    }
    get network() {
        return this.nftCollectionConfig.network;
    }
    get nftCollectionId() {
        return this.nftCollectionConfig.nftCollectionId;
    }
    getChain() {
        return this.nftCollectionConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Vet NFT Collection';
    }
    getBaseFactor() {
        return Math.pow(10, this.nftCollectionConfig.decimalPlaces);
    }
}
exports.VetNFTCollection = VetNFTCollection;
