// @bitgo/utxo-lib/dist/src/bitgo/PsbtUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProprietaryKeySubtype = exports.PSBT_PROPRIETARY_IDENTIFIER = void 0;
exports.getPsbtInputProprietaryKeyVals = getPsbtInputProprietaryKeyVals;
exports.getPsbtOutputProprietaryKeyVals = getPsbtOutputProprietaryKeyVals;
exports.getPsbtInputSignatureCount = getPsbtInputSignatureCount;
exports.isPsbtInputFinalized = isPsbtInputFinalized;
exports.isPsbt = isPsbt;
exports.toPsbtBuffer = toPsbtBuffer;
exports.withUnsafeNonSegwit = withUnsafeNonSegwit;
const ProprietaryKeyValUtils_1 = require("./ProprietaryKeyValUtils");
/**
 * bitgo proprietary key identifier
 */
exports.PSBT_PROPRIETARY_IDENTIFIER = 'BITGO';
/**
 * subtype for proprietary keys that bitgo uses
 */
var ProprietaryKeySubtype;
(function (ProprietaryKeySubtype) {
    ProprietaryKeySubtype[ProprietaryKeySubtype["ZEC_CONSENSUS_BRANCH_ID"] = 0] = "ZEC_CONSENSUS_BRANCH_ID";
    ProprietaryKeySubtype[ProprietaryKeySubtype["MUSIG2_PARTICIPANT_PUB_KEYS"] = 1] = "MUSIG2_PARTICIPANT_PUB_KEYS";
    ProprietaryKeySubtype[ProprietaryKeySubtype["MUSIG2_PUB_NONCE"] = 2] = "MUSIG2_PUB_NONCE";
    ProprietaryKeySubtype[ProprietaryKeySubtype["MUSIG2_PARTIAL_SIG"] = 3] = "MUSIG2_PARTIAL_SIG";
    ProprietaryKeySubtype[ProprietaryKeySubtype["PAYGO_ADDRESS_ATTESTATION_PROOF"] = 4] = "PAYGO_ADDRESS_ATTESTATION_PROOF";
    ProprietaryKeySubtype[ProprietaryKeySubtype["BIP322_MESSAGE"] = 5] = "BIP322_MESSAGE";
})(ProprietaryKeySubtype || (exports.ProprietaryKeySubtype = ProprietaryKeySubtype = {}));
/**
 * Search any data from psbt proprietary key value against keydata.
 * Default identifierEncoding is utf-8 for identifier.
 */
function getPsbtInputProprietaryKeyVals(input, keySearch) {
    if (!input.unknownKeyVals?.length) {
        return [];
    }
    return (0, ProprietaryKeyValUtils_1.getProprietaryKeyValuesFromUnknownKeyValues)(input, keySearch);
}
function getPsbtOutputProprietaryKeyVals(output, keySearch) {
    if (!output.unknownKeyVals?.length) {
        return [];
    }
    return (0, ProprietaryKeyValUtils_1.getProprietaryKeyValuesFromUnknownKeyValues)(output, keySearch);
}
/**
 * @return partialSig/tapScriptSig/MUSIG2_PARTIAL_SIG count iff input is not finalized
 */
function getPsbtInputSignatureCount(input) {
    if (isPsbtInputFinalized(input)) {
        throw new Error('Input is already finalized');
    }
    return Math.max(Array.isArray(input.partialSig) ? input.partialSig.length : 0, Array.isArray(input.tapScriptSig) ? input.tapScriptSig.length : 0, getPsbtInputProprietaryKeyVals(input, {
        identifier: exports.PSBT_PROPRIETARY_IDENTIFIER,
        subtype: ProprietaryKeySubtype.MUSIG2_PARTIAL_SIG,
    }).length);
}
/**
 * @return true iff PSBT input is finalized
 */
function isPsbtInputFinalized(input) {
    return Buffer.isBuffer(input.finalScriptSig) || Buffer.isBuffer(input.finalScriptWitness);
}
/**
 * @return true iff data starts with magic PSBT byte sequence
 * @param data byte array or hex string
 * */
function isPsbt(data) {
    // https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#specification
    // 0x70736274 - ASCII for 'psbt'. 0xff - separator
    if (typeof data === 'string') {
        if (data.length < 10) {
            return false;
        }
        data = Buffer.from(data.slice(0, 10), 'hex');
    }
    return 5 <= data.length && data.readUInt32BE(0) === 0x70736274 && data.readUInt8(4) === 0xff;
}
/**
 * First checks if the input is already a buffer that starts with the magic PSBT byte sequence.
 * If not, it checks if the input is a base64- or hex-encoded string that starts with PSBT header.
 *
 * This function is useful when reading a file that could be in any of the above formats or when
 * dealing with a request that could contain a hex or base64 encoded PSBT.
 *
 * @param data
 * @return buffer that starts with the magic PSBT byte sequence
 * @throws Error when conversion is not possible
 */
function toPsbtBuffer(data) {
    if (Buffer.isBuffer(data)) {
        // we are dealing with a buffer that looks like a psbt already
        if (isPsbt(data)) {
            return data;
        }
        // we could be dealing with a buffer that could be a hex or base64 encoded psbt
        data = data.toString('ascii');
    }
    if (typeof data === 'string') {
        const encodings = ['hex', 'base64'];
        for (const encoding of encodings) {
            let buffer;
            try {
                buffer = Buffer.from(data, encoding);
            }
            catch (e) {
                continue;
            }
            if (isPsbt(buffer)) {
                return buffer;
            }
        }
        throw new Error(`data is not in any of the following formats: ${encodings.join(', ')}`);
    }
    throw new Error('data must be a buffer or a string');
}
/**
 * This function allows signing or validating a psbt with non-segwit inputs those do not contain nonWitnessUtxo.
 */
function withUnsafeNonSegwit(psbt, fn, unsafe = true) {
    const prevValue = psbt.__CACHE.__UNSAFE_SIGN_NONSEGWIT;
    const prevValueWarn = psbt.__CACHE.__WARN_UNSAFE_SIGN_NONSEGWIT;
    psbt.__CACHE.__UNSAFE_SIGN_NONSEGWIT = unsafe;
    psbt.__CACHE.__WARN_UNSAFE_SIGN_NONSEGWIT = !unsafe;
    try {
        return fn();
    }
    finally {
        psbt.__CACHE.__UNSAFE_SIGN_NONSEGWIT = prevValue;
        psbt.__CACHE.__WARN_UNSAFE_SIGN_NONSEGWIT = prevValueWarn;
    }
}
