// @bitgo/sdk-coin-stx/dist/src/lib/fungibleTokenTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FungibleTokenTransferBuilder = void 0;
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactions_1 = require("@stacks/transactions");
const bn_js_1 = __importDefault(require("bn.js"));
const abstractContractBuilder_1 = require("./abstractContractBuilder");
const utils_1 = require("./utils");
const constants_1 = require("./constants");
class FungibleTokenTransferBuilder extends abstractContractBuilder_1.AbstractContractBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        this._fungibleTokenTransferParams = (0, utils_1.functionArgsToTokenTransferParams)(tx.stxTransaction.payload.functionArgs);
        this.contractAddress(this._contractAddress);
        this.contractName(this._contractName);
        this.functionName(this._functionName);
        this.functionArgs(this._functionArgs);
    }
    /** @inheritdoc */
    async buildImplementation() {
        this._postConditionMode = transactions_1.PostConditionMode.Deny;
        this._postConditions = this.tokenTransferParamsToPostCondition(this._fungibleTokenTransferParams);
        await super.buildImplementation();
        this.transaction.setTransactionType(sdk_core_1.TransactionType.Send);
        return this.transaction;
    }
    /**
     * Function to check if a transaction is a fungible token contract call
     *
     * @param {ContractCallPayload} payload
     * @returns {Boolean}
     */
    static isFungibleTokenTransferContractCall(payload) {
        return constants_1.FUNCTION_NAME_TRANSFER === payload.functionName.content;
    }
    /**
     * Set the token name
     *
     * @param {String} tokenName name of the token (@define-fungible-token value)
     * @returns {FungibleTokenTransferBuilder} This token transfer builder
     */
    tokenName(tokenName) {
        this._tokenName = tokenName;
        return this;
    }
    /**
     * Validate contract address
     *
     * @param {String} address contract address
     * @returns {FungibleTokenTransferBuilder} This token transfer builder
     */
    contractAddress(address) {
        if (!(0, utils_1.isValidAddress)(address)) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid address');
        }
        this._contractAddress = address;
        return this;
    }
    /**
     * Validate contract name
     *
     * @param {String} name contract name
     * @returns {FungibleTokenTransferBuilder} This token transfer builder
     */
    contractName(name) {
        if (name.length === 0) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid name');
        }
        this._contractName = name;
        return this;
    }
    /**
     * Validate function name
     *
     * @param {String} name function name
     * @returns {FungibleTokenTransferBuilder} This token transfer builder
     */
    functionName(name) {
        if (name.length === 0) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid name');
        }
        if (!(0, utils_1.isValidContractFunctionName)(name)) {
            throw new sdk_core_1.InvalidParameterValueError(`${name} is not supported contract function name`);
        }
        this._functionName = name;
        return this;
    }
    /**
     * Validate function arguments
     *
     * @param {ClarityValue[]} args array of clarity value as arguments
     * @returns {FungibleTokenTransferBuilder} This token transfer builder
     */
    functionArgs(args) {
        if (args.length < 3) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid number of arguments');
        }
        this._functionArgs = args;
        this._fungibleTokenTransferParams = (0, utils_1.functionArgsToTokenTransferParams)(this._functionArgs);
        return this;
    }
    /**
     * Function to convert token transfer params to post condition
     *
     * @param {TokenTransferParams} tokenTransferParams
     * @returns {PostCondition[]} returns stx fungible post condition
     */
    tokenTransferParamsToPostCondition(tokenTransferParams) {
        if (!this._tokenName) {
            const contractTokenName = (0, utils_1.findContractTokenNameUsingContract)(this._contractAddress, this._contractName);
            if (contractTokenName) {
                this.tokenName(contractTokenName);
            }
        }
        const amount = new bn_js_1.default(tokenTransferParams.amount);
        return [
            (0, transactions_1.makeStandardFungiblePostCondition)((0, utils_1.getSTXAddressFromPubKeys)(this._fromPubKeys, this._coinConfig.network.type === statics_1.NetworkType.MAINNET
                ? transactions_1.AddressVersion.MainnetMultiSig
                : transactions_1.AddressVersion.TestnetMultiSig, this._fromPubKeys.length > 1 ? transactions_1.AddressHashMode.SerializeP2SH : transactions_1.AddressHashMode.SerializeP2PKH, this._numberSignatures).address, transactions_1.FungibleConditionCode.Equal, amount, `${this._contractAddress}.${this._contractName}::${this._tokenName}`),
        ];
    }
}
exports.FungibleTokenTransferBuilder = FungibleTokenTransferBuilder;
