// @bitgo/sdk-coin-zketh/dist/src/zkethToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZkethToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class ZkethToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, ZkethToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, ZkethToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(ZkethToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /**
     * Make a query to Zksync explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response from Zksync explorer
     */
    async recoveryBlockchainExplorerQuery(query) {
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].zksyncExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl);
    }
    getFullName() {
        return 'Zketh Token';
    }
}
exports.ZkethToken = ZkethToken;
ZkethToken.coinNames = {
    Mainnet: 'zketh',
    Testnet: 'tzketh',
};
