// @bitgo/abstract-utxo/dist/cjs/test/unit/address.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../src");
const util_1 = require("./util");
const keychains = util_1.keychainsBase58.map((k) => ({ pub: k.pub }));
describe('assertFixedScriptWalletAddress', function () {
    describe('input validation', function () {
        it('throws InvalidAddressDerivationPropertyError when both chain and index are undefined', function () {
            assert.throws(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: undefined,
                index: undefined,
                keychains,
                format: 'base58',
                address: 'anything',
            }), sdk_core_1.InvalidAddressDerivationPropertyError);
        });
        it('throws InvalidAddressDerivationPropertyError when chain is non-finite', function () {
            assert.throws(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: Infinity,
                index: 0,
                keychains,
                format: 'base58',
                address: 'anything',
            }), sdk_core_1.InvalidAddressDerivationPropertyError);
        });
        it('throws InvalidAddressDerivationPropertyError when index is non-finite', function () {
            assert.throws(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 0,
                index: NaN,
                keychains,
                format: 'base58',
                address: 'anything',
            }), sdk_core_1.InvalidAddressDerivationPropertyError);
        });
        it('throws when keychains is missing', function () {
            assert.throws(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 0,
                index: 0,
                keychains: undefined,
                format: 'base58',
                address: 'anything',
            }), /missing required param keychains/);
        });
    });
    describe('address matching', function () {
        it('throws UnexpectedAddressError when address does not match derived address', function () {
            assert.throws(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 0,
                index: 0,
                keychains,
                format: 'base58',
                address: '3AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
            }), sdk_core_1.UnexpectedAddressError);
        });
        it('succeeds for btc p2sh (chain 0)', function () {
            const address = (0, src_1.generateAddress)('btc', { keychains, chain: 0 });
            assert.doesNotThrow(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 0,
                index: 0,
                keychains,
                format: 'base58',
                address,
            }));
        });
        it('succeeds for btc p2wsh (chain 20)', function () {
            const address = (0, src_1.generateAddress)('btc', { keychains, chain: 20 });
            assert.doesNotThrow(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 20,
                index: 0,
                keychains,
                format: 'base58',
                address,
            }));
        });
        it('succeeds for btc p2tr (chain 40)', function () {
            const address = (0, src_1.generateAddress)('btc', { keychains, chain: 40 });
            assert.doesNotThrow(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 40,
                index: 0,
                keychains,
                format: 'base58',
                address,
            }));
        });
        it('succeeds for bch cashaddr (chain 0)', function () {
            const address = (0, src_1.generateAddress)('bch', { keychains, chain: 0, format: 'cashaddr' });
            assert.doesNotThrow(() => (0, src_1.assertFixedScriptWalletAddress)('bch', {
                chain: 0,
                index: 0,
                keychains,
                format: 'cashaddr',
                address,
            }));
        });
        it('succeeds for a non-zero index', function () {
            const address = (0, src_1.generateAddress)('btc', { keychains, chain: 0, index: 5 });
            assert.doesNotThrow(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 0,
                index: 5,
                keychains,
                format: 'base58',
                address,
            }));
        });
        it('throws UnexpectedAddressError when index does not match', function () {
            const address = (0, src_1.generateAddress)('btc', { keychains, chain: 0, index: 0 });
            assert.throws(() => (0, src_1.assertFixedScriptWalletAddress)('btc', {
                chain: 0,
                index: 1,
                keychains,
                format: 'base58',
                address,
            }), sdk_core_1.UnexpectedAddressError);
        });
    });
});
