// @bitgo/abstract-utxo/dist/cjs/test/unit/bip322.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_core_1 = require("@bitgo/utxo-core");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const bip322_1 = require("../../src/transaction/bip322");
function createTestWalletKeys(seed) {
    const keys = utxolib.testutil.getKeyTriple(seed);
    return {
        xpubs: keys.map((k) => k.neutered().toBase58()),
        xprivs: keys.map((k) => k.toBase58()),
    };
}
function getDerivedPubkeys(seed, chain, index) {
    const keys = utxolib.testutil.getKeyTriple(seed);
    return keys.map((k) => k.derivePath(`m/0/0/${chain}/${index}`).publicKey.toString('hex'));
}
function getAddress(walletKeys, chain, index) {
    return wasm_utxo_1.fixedScriptWallet.address(walletKeys, chain, index, utxolib.networks.bitcoin);
}
describe('BIP322', function () {
    describe('BIP322MessageBroadcastable', () => {
        it('should serialize and deserialize correctly', () => {
            const message = {
                txHex: '010203',
                messageInfo: [
                    {
                        address: 'someAddress',
                        message: 'someMessage',
                        pubkeys: ['pubkey1', 'pubkey2', 'pubkey3'],
                        scriptType: 'p2sh',
                    },
                ],
            };
            const serialized = (0, bip322_1.serializeBIP322BroadcastableMessage)(message);
            const deserialized = (0, bip322_1.deserializeBIP322BroadcastableMessage)(serialized);
            assert_1.default.deepStrictEqual(deserialized, message);
        });
        it('should fail if there is an unsupported script type', function () {
            const message = {
                txHex: '010203',
                messageInfo: [
                    {
                        address: 'someAddress',
                        message: 'someMessage',
                        pubkeys: ['pubkey1', 'pubkey2', 'pubkey3'],
                        scriptType: 'unsupported',
                    },
                ],
            };
            const serialized = (0, bip322_1.serializeBIP322BroadcastableMessage)(message);
            assert_1.default.throws(() => {
                (0, bip322_1.deserializeBIP322BroadcastableMessage)(serialized);
            });
        });
    });
    describe('verifyTransactionFromBroadcastableMessage', function () {
        const seed = 'bip322_verify_test';
        const { xpubs, xprivs } = createTestWalletKeys(seed);
        const walletKeys = wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from(xpubs);
        it('should verify a valid p2shP2wsh PSBT proof', function () {
            const chain = 10;
            const index = 0;
            const messageText = 'Hello, BitGo!';
            const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            // Sign with user and bitgo keys
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastable = {
                txHex: Buffer.from(psbt.serialize()).toString('hex'),
                messageInfo: [
                    {
                        address,
                        message: messageText,
                        pubkeys,
                        scriptType: 'p2shP2wsh',
                    },
                ],
            };
            assert_1.default.strictEqual((0, bip322_1.verifyTransactionFromBroadcastableMessage)(broadcastable, 'btc'), true);
        });
        it('should verify a valid p2wsh PSBT proof', function () {
            const chain = 20;
            const index = 5;
            const messageText = 'P2WSH proof test';
            const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastable = {
                txHex: Buffer.from(psbt.serialize()).toString('hex'),
                messageInfo: [
                    {
                        address,
                        message: messageText,
                        pubkeys,
                        scriptType: 'p2wsh',
                    },
                ],
            };
            assert_1.default.strictEqual((0, bip322_1.verifyTransactionFromBroadcastableMessage)(broadcastable, 'btc'), true);
        });
        it('should verify multiple inputs in a single PSBT', function () {
            const messages = ['Message 1', 'Message 2'];
            const scriptIds = [
                { chain: 10, index: 0 },
                { chain: 20, index: 1 },
            ];
            const scriptTypes = ['p2shP2wsh', 'p2wsh'];
            const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            for (let i = 0; i < messages.length; i++) {
                wasm_utxo_1.bip322.addBip322Input(psbt, {
                    message: messages[i],
                    scriptId: scriptIds[i],
                    rootWalletKeys: walletKeys,
                });
            }
            // Sign all inputs
            for (let i = 0; i < messages.length; i++) {
                psbt.sign(i, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
                psbt.sign(i, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            }
            const messageInfo = messages.map((msg, i) => {
                const pubkeys = getDerivedPubkeys(seed, scriptIds[i].chain, scriptIds[i].index);
                return {
                    address: getAddress(walletKeys, scriptIds[i].chain, scriptIds[i].index),
                    message: msg,
                    pubkeys,
                    scriptType: scriptTypes[i],
                };
            });
            const broadcastable = {
                txHex: Buffer.from(psbt.serialize()).toString('hex'),
                messageInfo,
            };
            assert_1.default.strictEqual((0, bip322_1.verifyTransactionFromBroadcastableMessage)(broadcastable, 'btc'), true);
        });
        it('should return false for wrong message', function () {
            const chain = 10;
            const index = 0;
            const messageText = 'Original message';
            const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastable = {
                txHex: Buffer.from(psbt.serialize()).toString('hex'),
                messageInfo: [
                    {
                        address,
                        message: 'Different message', // Wrong message
                        pubkeys,
                        scriptType: 'p2shP2wsh',
                    },
                ],
            };
            assert_1.default.strictEqual((0, bip322_1.verifyTransactionFromBroadcastableMessage)(broadcastable, 'btc'), false);
        });
        it('should throw for unsupported coin', function () {
            const broadcastable = {
                txHex: '00',
                messageInfo: [],
            };
            assert_1.default.throws(() => (0, bip322_1.verifyTransactionFromBroadcastableMessage)(broadcastable, 'ltc'), /Only tbtc4 or btc coinNames are supported/);
        });
    });
    describe('generateBIP322MessageListAndVerifyFromMessageBroadcastable', function () {
        const seed = 'bip322_generate_test';
        const { xpubs, xprivs } = createTestWalletKeys(seed);
        const walletKeys = wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from(xpubs);
        it('should generate message list from valid proofs', function () {
            const chain = 10;
            const index = 0;
            const messageText = 'Proof message';
            const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastables = [
                {
                    txHex: Buffer.from(psbt.serialize()).toString('hex'),
                    messageInfo: [
                        {
                            address,
                            message: messageText,
                            pubkeys,
                            scriptType: 'p2shP2wsh',
                        },
                    ],
                },
            ];
            const result = (0, bip322_1.generateBIP322MessageListAndVerifyFromMessageBroadcastable)(broadcastables, 'btc');
            assert_1.default.strictEqual(result.length, 1);
            assert_1.default.strictEqual(result[0].address, address);
            assert_1.default.strictEqual(result[0].message, messageText);
        });
        it('should deduplicate addresses with same message', function () {
            const chain = 10;
            const index = 0;
            const messageText = 'Same message';
            // Create two separate PSBTs for the same address/message
            const psbt1 = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt1, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt1.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt1.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const psbt2 = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt2, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt2.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt2.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastables = [
                {
                    txHex: Buffer.from(psbt1.serialize()).toString('hex'),
                    messageInfo: [{ address, message: messageText, pubkeys, scriptType: 'p2shP2wsh' }],
                },
                {
                    txHex: Buffer.from(psbt2.serialize()).toString('hex'),
                    messageInfo: [{ address, message: messageText, pubkeys, scriptType: 'p2shP2wsh' }],
                },
            ];
            const result = (0, bip322_1.generateBIP322MessageListAndVerifyFromMessageBroadcastable)(broadcastables, 'btc');
            // Should deduplicate to a single entry
            assert_1.default.strictEqual(result.length, 1);
            assert_1.default.strictEqual(result[0].address, address);
            assert_1.default.strictEqual(result[0].message, messageText);
        });
        it('should throw for duplicate address with different message', function () {
            const chain = 10;
            const index = 0;
            const psbt1 = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt1, {
                message: 'Message 1',
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt1.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt1.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const psbt2 = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt2, {
                message: 'Message 2',
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt2.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt2.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastables = [
                {
                    txHex: Buffer.from(psbt1.serialize()).toString('hex'),
                    messageInfo: [{ address, message: 'Message 1', pubkeys, scriptType: 'p2shP2wsh' }],
                },
                {
                    txHex: Buffer.from(psbt2.serialize()).toString('hex'),
                    messageInfo: [{ address, message: 'Message 2', pubkeys, scriptType: 'p2shP2wsh' }],
                },
            ];
            assert_1.default.throws(() => (0, bip322_1.generateBIP322MessageListAndVerifyFromMessageBroadcastable)(broadcastables, 'btc'), /Duplicate address.*has different messages/);
        });
        it('should throw for invalid proof', function () {
            const chain = 10;
            const index = 0;
            const messageText = 'Valid message';
            const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.createEmpty('btc', walletKeys, { version: 0 });
            wasm_utxo_1.bip322.addBip322Input(psbt, {
                message: messageText,
                scriptId: { chain, index },
                rootWalletKeys: walletKeys,
            });
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[0]));
            psbt.sign(0, wasm_utxo_1.BIP32.fromBase58(xprivs[2]));
            const pubkeys = getDerivedPubkeys(seed, chain, index);
            const address = getAddress(walletKeys, chain, index);
            const broadcastables = [
                {
                    txHex: Buffer.from(psbt.serialize()).toString('hex'),
                    messageInfo: [
                        {
                            address,
                            message: 'Wrong message', // Doesn't match what was signed
                            pubkeys,
                            scriptType: 'p2shP2wsh',
                        },
                    ],
                },
            ];
            assert_1.default.throws(() => (0, bip322_1.generateBIP322MessageListAndVerifyFromMessageBroadcastable)(broadcastables, 'btc'), /did not have a successful validation/);
        });
    });
    describe('utxolib verification stack - wasm-utxo respects input.sighashType', function () {
        // This test verifies that wasm-utxo correctly respects the input.sighashType field
        // when creating musig2 partial signatures.
        //
        // Previously (before fix), wasm-utxo would always create signatures with SIGHASH_DEFAULT (0)
        // regardless of the input.sighashType field, causing validation to fail.
        //
        // Now (after fix), wasm-utxo reads input.sighashType and creates signatures with the
        // correct sighash type, allowing validation to succeed.
        it('should validate signatures when wasm-utxo respects input.sighashType', function () {
            const seed = 'p2trMusig2_sighash_test';
            const { xprivs } = createTestWalletKeys(seed);
            // Create utxolib RootWalletKeys for utxo-core PSBT construction
            const utxolibRootWalletKeys = new utxolib.bitgo.RootWalletKeys(utxolib.testutil.getKeyTriple(seed));
            // p2trMusig2 external chain code
            const chain = utxolib.bitgo.getExternalChainCode('p2trMusig2');
            const index = 0;
            const messageText = 'BIP322 sighash test';
            // Create BIP322 PSBT using utxo-core
            const psbt = utxo_core_1.bip322.createBaseToSignPsbt(utxolibRootWalletKeys, utxolib.networks.bitcoin);
            utxo_core_1.bip322.addBip322InputWithChainAndIndex(psbt, messageText, utxolibRootWalletKeys, { chain, index });
            // Note: utxo-core sets sighashType: Transaction.SIGHASH_ALL (1) for BIP322 inputs
            const SIGHASH_ALL = 1;
            assert_1.default.strictEqual(psbt.data.inputs[0].sighashType, SIGHASH_ALL);
            // Convert to wasm-utxo PSBT for cosigning
            const wasmPsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(psbt.toBuffer(), 'btc');
            // Generate musig2 nonces and sign with wasm-utxo
            // wasm-utxo now respects input.sighashType and creates signatures with SIGHASH_ALL
            const userKey = wasm_utxo_1.BIP32.fromBase58(xprivs[0]);
            const bitgoKey = wasm_utxo_1.BIP32.fromBase58(xprivs[2]);
            wasmPsbt.generateMusig2Nonces(userKey);
            wasmPsbt.generateMusig2Nonces(bitgoKey);
            wasmPsbt.sign(0, userKey);
            wasmPsbt.sign(0, bitgoKey);
            // Convert back to utxolib PSBT for validation
            const signedPsbt = utxolib.bitgo.createPsbtFromBuffer(Buffer.from(wasmPsbt.serialize()), utxolib.networks.bitcoin);
            // Validation should succeed because wasm-utxo now creates signatures
            // with the correct sighash type (SIGHASH_ALL) matching input.sighashType
            const validationResult = utxolib.bitgo.getSignatureValidationArrayPsbt(signedPsbt, utxolibRootWalletKeys);
            // Verify that both user (index 0) and bitgo (index 2) signatures are valid
            assert_1.default.strictEqual(validationResult.length, 1);
            const [inputIndex, sigValidation] = validationResult[0];
            assert_1.default.strictEqual(inputIndex, 0);
            assert_1.default.strictEqual(sigValidation[0], true, 'user signature should be valid');
            assert_1.default.strictEqual(sigValidation[1], false, 'backup signature should not be present');
            assert_1.default.strictEqual(sigValidation[2], true, 'bitgo signature should be valid');
        });
    });
});
