// @bitgo/abstract-utxo/dist/cjs/test/unit/customSigner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const utxoLib = __importStar(require("@bitgo/utxo-lib"));
require("should");
require("should-sinon");
const nock = require("nock");
const sinon = __importStar(require("sinon"));
const sdk_core_1 = require("@bitgo/sdk-core");
const util_1 = require("./util");
nock.disableNetConnect();
const bgUrl = sdk_core_1.common.Environments[util_1.defaultBitGo.getEnv()].uri;
const basecoin = (0, util_1.getUtxoCoin)('tbtc');
const wallet = (0, util_1.getUtxoWallet)(basecoin, {
    id: '5b34252f1bf349930e34020a',
    coin: 'tbtc',
    keys: ['5b3424f91bf349930e340175', '5b3424f91bf349930e340176', '5b3424f91bf349930e340177'],
});
describe('UTXO Custom Signer Function', function () {
    const recipients = [
        { address: 'abc', amount: 123 },
        { address: 'def', amount: 456 },
    ];
    const rootWalletKey = (0, util_1.getDefaultWalletKeys)();
    let customSigningFunction;
    let stubs;
    beforeEach(function () {
        customSigningFunction = sinon.stub().returns({
            txHex: 'this-is-a-tx',
        });
        stubs = [
            sinon.stub(wallet.baseCoin, 'postProcessPrebuild').returnsArg(0),
            sinon.stub(wallet.baseCoin, 'verifyTransaction').resolves(true),
            sinon.stub(wallet.baseCoin, 'signTransaction').resolves({ txHex: 'this-is-a-tx' }),
        ];
    });
    afterEach(function () {
        stubs.forEach((s) => s.restore());
        nock.cleanAll();
    });
    function nocks(txPrebuild) {
        const pubs = rootWalletKey.triple.map((k) => k.neutered().toBase58());
        nock(bgUrl)
            .post(`/api/v2/${wallet.coin()}/wallet/${wallet.id()}/tx/build`)
            .reply(200, { ...txPrebuild, txInfo: {} });
        nock(bgUrl).get(`/api/v2/${wallet.coin()}/public/block/latest`).reply(200, { height: 1000 });
        nock(bgUrl).persist().get(`/api/v2/${wallet.coin()}/key/${wallet.keyIds()[0]}`).reply(200, { pub: pubs[0] });
        nock(bgUrl).persist().get(`/api/v2/${wallet.coin()}/key/${wallet.keyIds()[1]}`).reply(200, { pub: pubs[1] });
        nock(bgUrl).persist().get(`/api/v2/${wallet.coin()}/key/${wallet.keyIds()[2]}`).reply(200, { pub: pubs[2] });
        return nock(bgUrl).post(`/api/v2/${wallet.coin()}/wallet/${wallet.id()}/tx/send`).reply(200, { ok: true });
    }
    it('should use a custom signing function if provided for PSBT with taprootKeyPathSpend input', async function () {
        const psbt = utxoLib.testutil.constructPsbt([{ scriptType: 'taprootKeyPathSpend', value: BigInt(1000) }], [{ scriptType: 'p2sh', value: BigInt(900) }], basecoin.network, rootWalletKey, 'unsigned');
        const scope = nocks({ txHex: psbt.toHex() });
        const result = await wallet.sendMany({ recipients, customSigningFunction });
        result.should.have.property('ok', true);
        customSigningFunction.should.have.been.calledTwice();
        scope.done();
    });
    it('should use a custom signing function if provided for PSBT without taprootKeyPathSpend input', async function () {
        const psbt = utxoLib.testutil.constructPsbt([{ scriptType: 'p2wsh', value: BigInt(1000) }], [{ scriptType: 'p2sh', value: BigInt(900) }], basecoin.network, rootWalletKey, 'unsigned');
        const scope = nocks({ txHex: psbt.toHex() });
        const result = await wallet.sendMany({ recipients, customSigningFunction });
        result.should.have.property('ok', true);
        customSigningFunction.should.have.been.calledOnce();
        scope.done();
    });
    it('should use a custom signing function if provided for Tx without taprootKeyPathSpend input', async function () {
        const tx = utxoLib.testutil.constructTxnBuilder([{ scriptType: 'p2wsh', value: BigInt(1000) }], [{ scriptType: 'p2sh', value: BigInt(900) }], basecoin.network, rootWalletKey, 'unsigned');
        const scope = nocks({ txHex: tx.buildIncomplete().toHex() });
        const result = await wallet.sendMany({ recipients, customSigningFunction });
        result.should.have.property('ok', true);
        customSigningFunction.should.have.been.calledOnce();
        scope.done();
    });
});
