// @bitgo/abstract-utxo/dist/cjs/test/unit/customChangeWalletFixtures.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const verifyKey_1 = require("../../src/verifyKey");
const KEY_ROLES = ['user', 'backup', 'bitgo'];
function signKeySignature(signerPrivateKey, pubToSign) {
    (0, assert_1.default)(signerPrivateKey.privateKey, 'signer must have a private key');
    return Buffer.from(wasm_utxo_1.message.signMessage(pubToSign, signerPrivateKey.privateKey)).toString('hex');
}
function createMockCustomChangeKeySignatures(signerPrivateKey, changeKeychains) {
    return {
        user: signKeySignature(signerPrivateKey, changeKeychains[0].pub),
        backup: signKeySignature(signerPrivateKey, changeKeychains[1].pub),
        bitgo: signKeySignature(signerPrivateKey, changeKeychains[2].pub),
    };
}
describe('Custom Change Wallet Key Signatures', function () {
    const mainUserKey = wasm_utxo_1.BIP32.fromSeedSha256('main-user');
    const changeKeys = [
        wasm_utxo_1.BIP32.fromSeedSha256('change-user'),
        wasm_utxo_1.BIP32.fromSeedSha256('change-backup'),
        wasm_utxo_1.BIP32.fromSeedSha256('change-bitgo'),
    ];
    const changeKeychains = changeKeys.map((k) => ({ pub: k.neutered().toBase58() }));
    it('should accept signatures from the correct key', function () {
        const signatures = createMockCustomChangeKeySignatures(mainUserKey, changeKeychains);
        for (const role of KEY_ROLES) {
            const index = KEY_ROLES.indexOf(role);
            assert_1.default.ok((0, verifyKey_1.verifyKeySignature)({
                userKeychain: { pub: mainUserKey.neutered().toBase58() },
                keychainToVerify: changeKeychains[index],
                keySignature: signatures[role],
            }), `failed to verify mock custom change ${role} key signature`);
        }
    });
    it('should reject signatures from a different key', function () {
        const wrongKey = wasm_utxo_1.BIP32.fromSeedSha256('wrong-key');
        const badSignatures = createMockCustomChangeKeySignatures(wrongKey, changeKeychains);
        for (const role of KEY_ROLES) {
            const index = KEY_ROLES.indexOf(role);
            assert_1.default.strictEqual((0, verifyKey_1.verifyKeySignature)({
                userKeychain: { pub: mainUserKey.neutered().toBase58() },
                keychainToVerify: changeKeychains[index],
                keySignature: badSignatures[role],
            }), false, `should have rejected custom change ${role} key signature from wrong key`);
        }
    });
});
