// @bitgo/abstract-utxo/dist/cjs/test/unit/customChangeWallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const strict_1 = __importDefault(require("node:assert/strict"));
const nock = require("nock");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const fixedScript_1 = require("../../src/transaction/fixedScript");
const verifyKey_1 = require("../../src/verifyKey");
const util_1 = require("./util");
function explainPsbt(psbt, walletKeys, customChangeWalletKeys, coin) {
    if (psbt instanceof wasm_utxo_1.fixedScriptWallet.BitGoPsbt) {
        return (0, fixedScript_1.explainPsbtWasm)(psbt, wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from(walletKeys), {
            replayProtection: { publicKeys: [] },
            customChangeWalletXpubs: customChangeWalletKeys
                ? wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from(customChangeWalletKeys)
                : undefined,
        });
    }
    else {
        return (0, fixedScript_1.explainPsbt)(psbt, { pubs: walletKeys, customChangePubs: customChangeWalletKeys }, coin);
    }
}
function describeWithBackend(sdkBackend) {
    describe(`Custom Change Wallets (sdkBackend=${sdkBackend})`, function () {
        const coin = (0, util_1.getUtxoCoin)('btc');
        const network = utxolib.networks.bitcoin;
        const bgUrl = sdk_core_1.common.Environments[util_1.defaultBitGo.getEnv()].uri;
        const rootWalletKeys = utxo_lib_1.testutil.getDefaultWalletKeys();
        const customChangeWalletKeys = utxo_lib_1.testutil.getWalletKeysForSeed('custom change');
        const userPrivateKey = wasm_utxo_1.BIP32.fromBase58(rootWalletKeys.triple[0].toBase58()).privateKey;
        const mainKeyIds = rootWalletKeys.triple.map((k) => (0, sdk_test_1.getSeed)(k.neutered().toBase58()).toString('hex'));
        const customChangeKeyIds = customChangeWalletKeys.triple.map((k) => (0, sdk_test_1.getSeed)(k.neutered().toBase58()).toString('hex'));
        const customChangeKeySignatures = Object.fromEntries(['user', 'backup', 'bitgo'].map((name, i) => [
            name,
            Buffer.from(wasm_utxo_1.message.signMessage(customChangeWalletKeys.triple[i].neutered().toBase58(), userPrivateKey)).toString('hex'),
        ]));
        const inputs = [{ scriptType: 'p2sh', value: BigInt(10000) }];
        const outputs = [
            // regular change (uses rootWalletKeys via default)
            { scriptType: 'p2sh', value: BigInt(3000) },
            // custom change (bip32Derivation from customChangeWalletKeys, not added as global xpubs)
            { scriptType: 'p2sh', value: BigInt(3000), walletKeys: customChangeWalletKeys },
            // external (no derivation info)
            { scriptType: 'p2sh', value: BigInt(3000), walletKeys: null },
        ];
        const utxolibPsbt = utxo_lib_1.testutil.constructPsbt(inputs, outputs, network, rootWalletKeys, 'unsigned', {
            addGlobalXPubs: true,
        });
        const psbt = sdkBackend === 'wasm-utxo' ? wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(utxolibPsbt.toBuffer(), 'btc') : utxolibPsbt;
        const externalAddress = utxolib.address.fromOutputScript(utxolibPsbt.txOutputs[2].script, network);
        const customChangeWalletId = 'custom-change-wallet-id';
        const mainWalletId = 'main-wallet-id';
        function nockKeyFetch(keyIds, keys) {
            return keyIds.map((id, i) => nock(bgUrl)
                .get(`/api/v2/${coin.getChain()}/key/${id}`)
                .reply(200, { pub: keys.triple[i].neutered().toBase58() }));
        }
        function nockCustomChangeWallet() {
            return nock(bgUrl).get(`/api/v2/${coin.getChain()}/wallet/${customChangeWalletId}`).reply(200, {
                id: customChangeWalletId,
                keys: customChangeKeyIds,
                coin: coin.getChain(),
            });
        }
        afterEach(() => nock.cleanAll());
        it('classifies custom change output when customChangePubs is provided', function () {
            const explanation = explainPsbt(psbt, rootWalletKeys, customChangeWalletKeys, 'btc');
            strict_1.default.strictEqual(explanation.changeOutputs.length, 1);
            strict_1.default.strictEqual(explanation.changeOutputs[0].amount, '3000');
            strict_1.default.ok(explanation.customChangeOutputs);
            strict_1.default.strictEqual(explanation.customChangeOutputs.length, 1);
            strict_1.default.strictEqual(explanation.customChangeOutputs[0].amount, '3000');
            strict_1.default.strictEqual(explanation.customChangeAmount, '3000');
            strict_1.default.strictEqual(explanation.outputs.length, 1);
            strict_1.default.strictEqual(explanation.outputs[0].amount, '3000');
        });
        it('classifies custom change output as external without customChangePubs', function () {
            const explanation = explainPsbt(psbt, rootWalletKeys, undefined, 'btc');
            strict_1.default.strictEqual(explanation.changeOutputs.length, 1);
            strict_1.default.strictEqual(explanation.changeOutputs[0].amount, '3000');
            strict_1.default.strictEqual(explanation.customChangeOutputs?.length ?? 0, 0);
            // custom change + external both treated as external outputs
            strict_1.default.strictEqual(explanation.outputs.length, 2);
        });
        it('verifies valid custom change key signatures', function () {
            const userPub = rootWalletKeys.triple[0].neutered().toBase58();
            for (const key of customChangeWalletKeys.triple) {
                const pub = key.neutered().toBase58();
                const signature = Buffer.from(wasm_utxo_1.message.signMessage(pub, userPrivateKey)).toString('hex');
                strict_1.default.ok((0, verifyKey_1.verifyKeySignature)({ userKeychain: { pub: userPub }, keychainToVerify: { pub }, keySignature: signature }));
            }
        });
        it('rejects invalid custom change key signatures', function () {
            const wrongKey = wasm_utxo_1.BIP32.fromBase58(utxo_lib_1.testutil.getWalletKeysForSeed('wrong').triple[0].toBase58());
            const userPub = rootWalletKeys.triple[0].neutered().toBase58();
            for (const key of customChangeWalletKeys.triple) {
                const pub = key.neutered().toBase58();
                const badSignature = Buffer.from(wasm_utxo_1.message.signMessage(pub, wrongKey.privateKey)).toString('hex');
                strict_1.default.strictEqual((0, verifyKey_1.verifyKeySignature)({ userKeychain: { pub: userPub }, keychainToVerify: { pub }, keySignature: badSignature }), false);
            }
        });
        describe('parseTransaction', function () {
            it('fetches custom change wallet keys and verifies signatures', async function () {
                const wallet = new sdk_core_1.Wallet(util_1.defaultBitGo, coin, {
                    id: mainWalletId,
                    keys: mainKeyIds,
                    coin: coin.getChain(),
                    coinSpecific: { customChangeWalletId },
                    customChangeKeySignatures,
                });
                const nocks = [
                    ...nockKeyFetch(mainKeyIds, rootWalletKeys),
                    nockCustomChangeWallet(),
                    ...nockKeyFetch(customChangeKeyIds, customChangeWalletKeys),
                ];
                const parsed = await coin.parseTransaction({
                    txParams: { recipients: [{ address: externalAddress, amount: '3000' }] },
                    txPrebuild: { txHex: utxolibPsbt.toHex(), decodeWith: sdkBackend },
                    wallet: wallet,
                });
                for (const n of nocks)
                    strict_1.default.ok(n.isDone());
                strict_1.default.ok(parsed.customChange);
                strict_1.default.strictEqual(parsed.customChange.keys.length, 3);
                for (let i = 0; i < 3; i++) {
                    strict_1.default.strictEqual(parsed.customChange.keys[i].pub, customChangeWalletKeys.triple[i].neutered().toBase58());
                }
                strict_1.default.strictEqual(parsed.explicitExternalOutputs.length, 1);
                strict_1.default.strictEqual(parsed.explicitExternalOutputs[0].amount, '3000');
            });
            it('has no custom change when wallet lacks customChangeWalletId', async function () {
                const wallet = new sdk_core_1.Wallet(util_1.defaultBitGo, coin, {
                    id: mainWalletId,
                    keys: mainKeyIds,
                    coin: coin.getChain(),
                    coinSpecific: {},
                });
                const nocks = nockKeyFetch(mainKeyIds, rootWalletKeys);
                const parsed = await coin.parseTransaction({
                    txParams: { recipients: [{ address: externalAddress, amount: '3000' }] },
                    txPrebuild: { txHex: utxolibPsbt.toHex(), decodeWith: sdkBackend },
                    wallet: wallet,
                });
                for (const n of nocks)
                    strict_1.default.ok(n.isDone());
                strict_1.default.strictEqual(parsed.customChange, undefined);
                strict_1.default.strictEqual(parsed.needsCustomChangeKeySignatureVerification, false);
            });
            it('rejects invalid custom change key signatures', async function () {
                const wrongKey = wasm_utxo_1.BIP32.fromBase58(utxo_lib_1.testutil.getWalletKeysForSeed('wrong').triple[0].toBase58());
                const badSignatures = Object.fromEntries(['user', 'backup', 'bitgo'].map((name, i) => [
                    name,
                    Buffer.from(wasm_utxo_1.message.signMessage(customChangeWalletKeys.triple[i].neutered().toBase58(), wrongKey.privateKey)).toString('hex'),
                ]));
                const wallet = new sdk_core_1.Wallet(util_1.defaultBitGo, coin, {
                    id: mainWalletId,
                    keys: mainKeyIds,
                    coin: coin.getChain(),
                    coinSpecific: { customChangeWalletId },
                    customChangeKeySignatures: badSignatures,
                });
                nockKeyFetch(mainKeyIds, rootWalletKeys);
                nockCustomChangeWallet();
                nockKeyFetch(customChangeKeyIds, customChangeWalletKeys);
                await strict_1.default.rejects(() => coin.parseTransaction({
                    txParams: { recipients: [{ address: externalAddress, amount: '3000' }] },
                    txPrebuild: { txHex: utxolibPsbt.toHex(), decodeWith: sdkBackend },
                    wallet: wallet,
                }), /failed to verify custom change .* key signature/);
            });
        });
    });
}
describeWithBackend('utxolib');
describeWithBackend('wasm-utxo');
