// @bitgo/abstract-utxo/dist/cjs/test/unit/descriptorAddress.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDescriptorAddress = getDescriptorAddress;
const assert = __importStar(require("node:assert"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const testutils = __importStar(require("@bitgo/wasm-utxo/testutils"));
const src_1 = require("../../src");
const util_1 = require("./util");
function getDescriptorAddress(d, index, coinName) {
    const derivedScript = src_1.descriptor.Descriptor.fromString(d, 'derivable').atDerivationIndex(index).scriptPubkey();
    return wasm_utxo_1.address.fromOutputScriptWithCoin(derivedScript, coinName);
}
describe('descriptor wallets', function () {
    const coin = (0, util_1.getUtxoCoin)('tbtc');
    const xpubs = testutils.getKeyTriple('setec astronomy').map((k) => k.neutered().toBase58());
    function withChecksum(descriptor) {
        return src_1.descriptor.Descriptor.fromString(descriptor, 'derivable').toString();
    }
    function getNamedDescriptor2Of2(name, a, b) {
        return {
            name,
            value: withChecksum(`sh(multi(2,${a}/*,${b}/*))`),
            signatures: [],
        };
    }
    function getIWalletWithDescriptors(descriptors) {
        return {
            coinSpecific() {
                return { descriptors };
            },
        };
    }
    const descFoo = getNamedDescriptor2Of2('foo', xpubs[0], xpubs[1]);
    const descBar = getNamedDescriptor2Of2('bar', xpubs[1], xpubs[0]);
    const addressFoo0 = getDescriptorAddress(descFoo.value, 0, coin.name);
    const addressFoo1 = getDescriptorAddress(descFoo.value, 1, coin.name);
    const addressBar0 = getDescriptorAddress(descBar.value, 0, coin.name);
    it('has expected values', function () {
        assert.deepStrictEqual([addressFoo0, addressFoo1, addressBar0], [
            '2N9b1trWxMJN16mTzGJypFn6pEWfXtgh689',
            '2N1YFzj4ECzcjuruaEvSzGaGGH1topMXMXZ',
            '2N9oN5Kc2fLt2MrxEkuQPsy8Fg2KdrFfeKH',
        ]);
    });
    function runTestIsAddress(address, index, descriptorName, descriptorChecksum, expected) {
        it(`should return ${expected} for address ${address} with index ${index} and descriptor ${descriptorName} with checksum ${descriptorChecksum}`, async function () {
            const wallet = getIWalletWithDescriptors([descFoo, descBar]);
            async function f() {
                return coin.isWalletAddress({
                    address,
                    index,
                    coinSpecific: { descriptorName, descriptorChecksum },
                    keychains: xpubs.map((pub) => ({ pub })),
                }, wallet);
            }
            if (expected === true) {
                assert.equal(await f(), expected);
            }
            else {
                // because isWalletAddress is stupid it actually throws instead of returning false
                await assert.rejects(f, expected);
            }
        });
    }
    runTestIsAddress(addressFoo0, 0, 'foo', descFoo.value.slice(-8), true);
    runTestIsAddress(addressFoo1, 0, 'foo', descFoo.value.slice(-8), /Address mismatch for descriptor/);
    runTestIsAddress(addressBar0, 0, 'bar', descFoo.value.slice(-8), /Descriptor checksum mismatch/);
    runTestIsAddress(addressFoo0, 0, 'bar', descBar.value.slice(-8), /Address mismatch for descriptor/);
});
