// @bitgo/abstract-utxo/dist/cjs/test/unit/explainTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const nock = require("nock");
const fixtures_1 = require("./fixtures/bip322/fixtures");
const psbtHexProof_1 = require("./fixtures/psbtHexProof");
const util_1 = require("./util");
nock.disableNetConnect();
const bgUrl = sdk_core_1.common.Environments[util_1.defaultBitGo.getEnv()].uri;
describe('Explain Transaction', function () {
    afterEach(function () {
        nock.cleanAll();
    });
    describe('Verify paygo output when explaining psbt transaction', function () {
        const coin = (0, util_1.getUtxoCoin)('tbtc4');
        const pubs = [
            'xpub661MyMwAqRbcFaKvNBFdV6HY7ibXxFSbL7rDjY1cVM8s3pGPTNKfjTu8SmatNZ7AcZQehSqcEnC7vezMoprQvhqQUszLhuY4G8ruv6PGEr7',
            'xpub661MyMwAqRbcGkAVVQVHrEYQA4hfbDW9Rpn35b6sXA9TSBd5Qzjwz7F6Weje57kBVeVfimfJjXutwUDBSMz5yRwsWik9gNyxrdvSaJbjgi6',
            'xpub661MyMwAqRbcGCCL3GYNbvKs1t5k5yeKZcV5smto9T5Z17zkcgRF4X9uzDfPxMHHedwF4JcJ6kpg8M2NWHEFC5LMSv1t3nMMm1GC9PcVmq5',
        ];
        const keyIds = ['key-user', 'key-backup', 'key-bitgo'];
        function nockKeyFetch() {
            keyIds.forEach((id, i) => {
                nock(bgUrl).get(`/api/v2/${coin.getChain()}/key/${id}`).reply(200, { pub: pubs[i] });
            });
        }
        it('should detect and verify paygo address proof in PSBT', async function () {
            nockKeyFetch();
            const wallet = new sdk_core_1.Wallet(util_1.defaultBitGo, coin, {
                id: 'mock-wallet-id',
                coin: coin.getChain(),
                keys: keyIds,
            });
            await coin.explainTransaction(psbtHexProof_1.psbtTxHex, wallet);
        });
    });
    describe('BIP322 Proof', function () {
        const coin = (0, util_1.getUtxoCoin)('btc');
        const pubs = fixtures_1.bip322Fixtures.valid.rootWalletKeys.triple.map((b) => b.neutered().toBase58());
        it('should successfully run with a user nonce', async function () {
            const psbtHex = fixtures_1.bip322Fixtures.valid.userNonce;
            const result = await coin.explainTransaction({ txHex: psbtHex, pubs });
            assert_1.default.strictEqual(result.outputAmount, '0');
            assert_1.default.strictEqual(result.changeAmount, '0');
            assert_1.default.strictEqual(result.outputs.length, 1);
            assert_1.default.strictEqual(result.outputs[0].address, 'scriptPubKey:6a');
            assert_1.default.strictEqual(result.fee, '0');
            assert_1.default.ok('signatures' in result);
            assert_1.default.strictEqual(result.signatures, 0);
            assert_1.default.ok(result.messages);
            result.messages?.forEach((obj) => {
                assert_1.default.ok(obj.address);
                assert_1.default.ok(obj.message);
                assert_1.default.strictEqual(obj.message, fixtures_1.bip322Fixtures.valid.message);
            });
        });
        it('should successfully run with a user signature', async function () {
            const psbtHex = fixtures_1.bip322Fixtures.valid.userSignature;
            const result = await coin.explainTransaction({ txHex: psbtHex, pubs });
            assert_1.default.strictEqual(result.outputAmount, '0');
            assert_1.default.strictEqual(result.changeAmount, '0');
            assert_1.default.strictEqual(result.outputs.length, 1);
            assert_1.default.strictEqual(result.outputs[0].address, 'scriptPubKey:6a');
            assert_1.default.strictEqual(result.fee, '0');
            assert_1.default.ok('signatures' in result);
            assert_1.default.strictEqual(result.signatures, 1);
            assert_1.default.ok(result.messages);
            result.messages?.forEach((obj) => {
                assert_1.default.ok(obj.address);
                assert_1.default.ok(obj.message);
                assert_1.default.strictEqual(obj.message, fixtures_1.bip322Fixtures.valid.message);
            });
        });
        it('should successfully run with a hsm signature', async function () {
            const psbtHex = fixtures_1.bip322Fixtures.valid.hsmSignature;
            const result = await coin.explainTransaction({ txHex: psbtHex, pubs });
            assert_1.default.strictEqual(result.outputAmount, '0');
            assert_1.default.strictEqual(result.changeAmount, '0');
            assert_1.default.strictEqual(result.outputs.length, 1);
            assert_1.default.strictEqual(result.outputs[0].address, 'scriptPubKey:6a');
            assert_1.default.strictEqual(result.fee, '0');
            assert_1.default.ok('signatures' in result);
            assert_1.default.strictEqual(result.signatures, 2);
            assert_1.default.ok(result.messages);
            result.messages?.forEach((obj) => {
                assert_1.default.ok(obj.address);
                assert_1.default.ok(obj.message);
                assert_1.default.strictEqual(obj.message, fixtures_1.bip322Fixtures.valid.message);
            });
        });
    });
});
