// @bitgo/abstract-utxo/dist/cjs/test/unit/keychains.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
require("should");
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const btc_1 = require("../../src/impl/btc");
const util_1 = require("./util");
function run(coin) {
    describe(`UTXO Keychains ${coin.getChain()}`, function () {
        it('validates pub', function () {
            const { pub } = coin.keychains().create();
            assert.ok(pub);
            coin.isValidPub(pub).should.equal(true);
        });
    });
}
util_1.utxoCoins.forEach((c) => run(c));
describe('Audit Key', function () {
    // Encrypted backup key fixture for testing assertIsValidKey
    const btcBackupKey = {
        key: '{"iv":"JgqqE4W45/tKBSMSYqD+qg==","v":1,"iter":10000,"ks":256,"ts":64,"mode"' +
            ':"ccm","adata":"","cipher":"aes","salt":"kiLPf8VSdI0=","ct":"zUh4Oko/06g02E' +
            'wnqOfzJbTwtE2p3b19jDk8Tum07Jv3N/RP7Bo0w/ObLBO1uIJFossO3nJ1JS+7t/vPQhdCtN8oD' +
            '6YrZnEZYrRwN6JQkL1uYPnZ1PoWbYI9navK5CLU1KQwDTO9YEN46++OrzFH+CjpQVLblaw="}',
    };
    let bitgo;
    let coin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
        bitgo.safeRegister('tbtc', btc_1.Tbtc.createInstance);
        bitgo.initializeTestVars();
        coin = bitgo.coin('tbtc');
    });
    it('should return for valid inputs', function () {
        coin.assertIsValidKey({
            encryptedPrv: btcBackupKey.key,
            walletPassphrase: 'kAm[EFQ6o=SxlcLFDw%,',
        });
    });
    it('should throw error if the walletPassphrase is incorrect', function () {
        assert.throws(() => coin.assertIsValidKey({
            encryptedPrv: btcBackupKey.key,
            walletPassphrase: 'foo',
        }), { message: "failed to decrypt prv: ccm: tag doesn't match" });
    });
    it('should throw if the key is altered', function () {
        const alteredKey = btcBackupKey.key.replace(/[0-9]/g, '0');
        assert.throws(() => coin.assertIsValidKey({
            encryptedPrv: alteredKey,
            walletPassphrase: 'kAm[EFQ6o=SxlcLFDw%,',
        }), { message: 'failed to decrypt prv: json decrypt: invalid parameters' });
    });
});
