// @bitgo/abstract-utxo/dist/cjs/test/unit/parseTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const sinon = __importStar(require("sinon"));
const sdk_core_1 = require("@bitgo/sdk-core");
const util_1 = require("./util");
describe('Parse Transaction', function () {
    const coin = (0, util_1.getUtxoCoin)('tbtc');
    /*
     * mock objects which get passed into parse transaction.
     * These objects are structured to force parse transaction into a
     * particular execution path for these tests.
     */
    const verification = {
        disableNetworking: true,
        keychains: {
            user: { id: '0', pub: 'aaa', type: 'independent' },
            backup: { id: '1', pub: 'bbb', type: 'independent' },
            bitgo: { id: '2', pub: 'ccc', type: 'independent' },
        },
    };
    const wallet = sinon.createStubInstance(sdk_core_1.Wallet, {
        migratedFrom: '2MzJxAENaesCFu3orrCdj22c69tLEsKXQoR',
    });
    const outputAmount = (0.01 * 1e8).toString();
    let stubExplainTransaction;
    let stubVerifyAddress;
    afterEach(() => {
        if (stubExplainTransaction) {
            stubExplainTransaction.restore();
        }
        if (stubVerifyAddress) {
            stubVerifyAddress.restore();
        }
    });
    async function runClassifyOutputsTest(outputAddress, verification, expectExternal, txParams = {}) {
        stubExplainTransaction = sinon.stub(coin, 'explainTransaction').resolves({
            outputs: [],
            changeOutputs: [
                {
                    address: outputAddress,
                    amount: outputAmount,
                },
            ],
        });
        if (!txParams.changeAddress) {
            stubVerifyAddress = sinon.stub(coin, 'verifyAddress').throws(new sdk_core_1.UnexpectedAddressError('test error'));
        }
        const parsedTransaction = await coin.parseTransaction({
            txParams,
            txPrebuild: { txHex: '' },
            wallet: wallet,
            verification,
        });
        assert_1.default.ok(parsedTransaction.outputs[0]);
        assert_1.default.deepStrictEqual(parsedTransaction.outputs[0], {
            address: outputAddress,
            amount: outputAmount,
            external: expectExternal,
        });
        const isExplicit = txParams.recipients !== undefined && txParams.recipients.some((recipient) => recipient.address === outputAddress);
        assert_1.default.strictEqual(parsedTransaction.explicitExternalSpendAmount, Number(isExplicit && expectExternal ? outputAmount : 0));
        assert_1.default.strictEqual(parsedTransaction.implicitExternalSpendAmount, Number(!isExplicit && expectExternal ? outputAmount : 0));
    }
    it('should classify outputs which spend change back to a v1 wallet base address as internal', async function () {
        return runClassifyOutputsTest(wallet.migratedFrom(), verification, false);
    });
    it('should classify outputs which spend change back to a v1 wallet base address as external ' +
        'if considerMigratedFromAddressInternal is set and false', async function () {
        return runClassifyOutputsTest(wallet.migratedFrom(), { ...verification, considerMigratedFromAddressInternal: false }, true);
    });
    it('should classify outputs which spend to addresses not on the wallet as external', async function () {
        return runClassifyOutputsTest('2Mxjx4E2EEe4yJuLvdEuAdMUd4id1emPCZs', verification, true);
    });
    it('should accept a custom change address', async function () {
        const changeAddress = '2NAuziD75WnPPHJVwnd4ckgY4SuJaDVVbMD';
        return runClassifyOutputsTest(changeAddress, verification, false, {
            changeAddress,
            recipients: [],
        });
    });
    it('should classify outputs with external address in recipients as explicit', async function () {
        const externalAddress = '2NAuziD75WnPPHJVwnd4ckgY4SuJaDVVbMD';
        return runClassifyOutputsTest(externalAddress, verification, true, {
            recipients: [{ address: externalAddress, amount: outputAmount }],
        });
    });
});
