// @bitgo/abstract-utxo/dist/cjs/test/unit/prebuildAndSign.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const nock = require("nock");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const util_1 = require("./util");
const walletPassphrase = 'gabagool';
const webauthnWalletPassPhrase = 'just the gabagool';
function assertSignable(psbtHex, inputScripts, network) {
    const psbt = utxolib.bitgo.createPsbtFromHex(psbtHex, network);
    // Make sure that you can sign with bitgo key and extract the transaction
    // Skip validation for p2shP2pk (single-sig replay protection) and taprootKeyPathSpend (requires musig2 nonce exchange)
    if (inputScripts.includes('p2shP2pk') || inputScripts.includes('taprootKeyPathSpend')) {
        return;
    }
    const key = inputScripts.includes('p2trMusig2') ? rootWalletKeys.backup : rootWalletKeys.bitgo;
    psbt.signAllInputsHD(key, { deterministic: true });
    psbt.validateSignaturesOfAllInputs();
    psbt.finalizeAllInputs();
    const tx = psbt.extractTransaction();
    assert.ok(tx);
}
// Build the key objects
const rootWalletKeys = (0, util_1.getDefaultWalletKeys)();
const keyDocumentObjects = rootWalletKeys.triple.map((bip32, keyIdx) => {
    return {
        id: (0, sdk_test_1.getSeed)(util_1.keychainsBase58[keyIdx].pub).toString('hex'),
        pub: bip32.neutered().toBase58(),
        source: ['user', 'backup', 'bitgo'][keyIdx],
        encryptedPrv: (0, util_1.encryptKeychain)(walletPassphrase, util_1.keychainsBase58[keyIdx]),
        webauthnDevices: [
            {
                otpDeviceId: '123',
                authenticatorInfo: {
                    credID: 'credID',
                    fmt: 'packed',
                    publicKey: 'some value',
                },
                prfSalt: '456',
                encryptedPrv: (0, util_1.encryptKeychain)(webauthnWalletPassPhrase, util_1.keychainsBase58[keyIdx]),
            },
        ],
        coinSpecific: {},
    };
});
function run(coin, inputScripts, txFormat) {
    function createPrebuildPsbt(inputs, outputs) {
        const psbt = utxolib.testutil.constructPsbt(inputs, outputs, coin.network, rootWalletKeys, 'unsigned');
        utxolib.bitgo.addXpubsToPsbt(psbt, rootWalletKeys);
        return psbt;
    }
    function createNocks(params) {
        const nocks = [];
        // Nock the prebuild route (/tx/build, blockheight)
        const expected_params = {
            recipients: [params.recipient],
            rbfTxIds: params.rbfTxIds,
            feeMultiplier: params.feeMultiplier,
            changeAddressType: ['p2trMusig2', 'p2wsh', 'p2shP2wsh', 'p2sh', 'p2tr'],
        };
        if (params.txFormat) {
            expected_params['txFormat'] = params.txFormat;
        }
        nocks.push(nock(params.bgUrl)
            .post(`/api/v2/${coin.getChain()}/wallet/${params.wallet.id()}/tx/build`, expected_params)
            .reply(200, { txHex: params.prebuild.toHex(), txInfo: {} }));
        nocks.push(nock(params.bgUrl).get(`/api/v2/${coin.getChain()}/public/block/latest`).reply(200, { height: 1000 }));
        // nock the keychain fetch - 3 times (prebuildAndSign, verifyTransaction, and signTransaction)
        params.keyDocuments.forEach((keyDocument) => {
            nocks.push(nock(params.bgUrl).get(`/api/v2/${coin.getChain()}/key/${keyDocument.id}`).times(3).reply(200, keyDocument));
        });
        // nock the address info fetch
        if (params.nockOutputAddresses) {
            nocks.push(nock(params.bgUrl)
                .get(`/api/v2/${coin.getChain()}/wallet/${params.wallet.id()}/address/${params.addressInfo.address}`)
                .reply(200, params.addressInfo));
        }
        if (params.rbfTxIds) {
            nocks.push(nock(params.bgUrl)
                .get(`/api/v2/${coin.getChain()}/wallet/${params.wallet.id()}/tx/${params.rbfTxIds[0]}?includeRbf=true`)
                .reply(200, {
                outputs: [
                    {
                        address: params.recipient.address,
                        value: Number(params.recipient.amount),
                        valueString: params.recipient.amount,
                        wallet: params.selfSend ? params.wallet.id() : 'some-other-wallet-id', // external output if not a self send
                    },
                    // Dummy change output to test transfer entries filtering
                    {
                        address: params.recipient.address,
                        value: Number(params.recipient.amount),
                        valueString: params.recipient.amount,
                        wallet: params.wallet.id(), // internal output
                    },
                ],
            }));
        }
        // nock the deterministic nonce response
        if (inputScripts.includes('taprootKeyPathSpend')) {
            const psbt = params.prebuild.clone();
            psbt.setAllInputsMusig2NonceHD(rootWalletKeys.user);
            psbt.setAllInputsMusig2NonceHD(rootWalletKeys.bitgo);
            nocks.push(nock(params.bgUrl)
                .post(`/api/v2/${coin.getChain()}/wallet/${params.wallet.id()}/tx/signpsbt`, (body) => body.psbt)
                .reply(200, { psbt: psbt.toHex() }));
        }
        return nocks;
    }
    describe(`${coin.getFullName()}-prebuildAndSign-txFormat=${txFormat}-inputScripts=${inputScripts.join(',')}`, function () {
        const wallet = (0, util_1.getUtxoWallet)(coin, {
            coinSpecific: { addressVersion: 'base58' },
            keys: keyDocumentObjects.map((k) => k.id),
            id: 'walletId',
        });
        const bgUrl = sdk_core_1.common.Environments[util_1.defaultBitGo.getEnv()].uri;
        let prebuild;
        let recipient;
        let addressInfo;
        const fee = BigInt(10000);
        before(async function () {
            // Make output address information
            const outputAmount = BigInt(inputScripts.length) * BigInt(1e8) - fee;
            const outputScriptType = 'p2sh';
            const outputChain = utxolib.bitgo.getExternalChainCode(outputScriptType);
            const outputAddress = utxolib.bitgo.getWalletAddress(rootWalletKeys, outputChain, 0, coin.network);
            recipient = {
                address: outputAddress,
                amount: outputAmount.toString(),
            };
            addressInfo = {
                address: outputAddress,
                chain: outputChain,
                index: 0,
                coin: coin.getChain(),
                wallet: wallet.id(),
                coinSpecific: {},
            };
            prebuild = createPrebuildPsbt(inputScripts.map((s) => ({ scriptType: s, value: BigInt(1e8) })), [{ scriptType: outputScriptType, value: outputAmount }]);
        });
        afterEach(nock.cleanAll);
        [true, false].forEach((useWebauthn) => {
            it(`should succeed with ${useWebauthn ? 'webauthn encryptedPrv' : 'encryptedPrv'}`, async function () {
                // Check if this wallet/coin combination defaults to psbt
                const defaultTxFormat = coin.getDefaultTxFormat(wallet);
                const nocks = createNocks({
                    bgUrl,
                    wallet,
                    keyDocuments: keyDocumentObjects,
                    prebuild,
                    recipient,
                    addressInfo,
                    nockOutputAddresses: txFormat !== 'psbt',
                    txFormat: defaultTxFormat,
                });
                // call prebuild and sign, nocks should be consumed
                const res = (await wallet.prebuildAndSignTransaction({
                    recipients: [recipient],
                    walletPassphrase: useWebauthn ? webauthnWalletPassPhrase : walletPassphrase,
                }));
                nocks.forEach((nock) => assert.ok(nock.isDone()));
                assertSignable(res.txHex, inputScripts, coin.network);
            });
            it('should fail if the wallet passphrase is incorrect', async function () {
                createNocks({
                    bgUrl,
                    wallet,
                    keyDocuments: keyDocumentObjects,
                    prebuild,
                    recipient,
                    addressInfo,
                    nockOutputAddresses: txFormat !== 'psbt',
                });
                await wallet
                    .prebuildAndSignTransaction({
                    recipients: [recipient],
                    walletPassphrase: Math.random().toString(),
                })
                    .should.be.rejectedWith('unable to decrypt keychain with the given wallet passphrase');
            });
        });
        [true, false].forEach((selfSend) => {
            it(`should be able to build, sign, & verify a replacement transaction with selfSend: ${selfSend}`, async function () {
                const rbfTxIds = ['tx-to-be-replaced'], feeMultiplier = 1.5;
                // Check if this wallet/coin combination defaults to psbt
                const defaultTxFormat = coin.getDefaultTxFormat(wallet);
                const nocks = createNocks({
                    bgUrl,
                    wallet,
                    keyDocuments: keyDocumentObjects,
                    prebuild,
                    recipient,
                    addressInfo,
                    rbfTxIds,
                    feeMultiplier,
                    selfSend,
                    nockOutputAddresses: txFormat !== 'psbt',
                    txFormat: defaultTxFormat,
                });
                // call prebuild and sign, nocks should be consumed
                const res = (await wallet.prebuildAndSignTransaction({
                    recipients: [recipient],
                    walletPassphrase,
                    rbfTxIds,
                    feeMultiplier,
                }));
                nocks.forEach((nock) => assert.ok(nock.isDone()));
                assertSignable(res.txHex, inputScripts, coin.network);
            });
        });
    });
}
(0, util_1.getMinUtxoCoins)().forEach((coin) => {
    run(coin, (0, util_1.getScriptTypes)(coin, 'psbt'), 'psbt');
});
