// @bitgo/abstract-utxo/dist/cjs/test/unit/signTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
require("mocha");
const assert = __importStar(require("assert"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const nock = require("nock");
const utxo_lib_1 = require("@bitgo/utxo-lib");
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../src");
const util_1 = require("./util");
describe('signTransaction', function () {
    const bgUrl = sdk_core_1.common.Environments[util_1.defaultBitGo.getEnv()].uri;
    const coin = (0, util_1.getUtxoCoin)('btc');
    const wallet = (0, util_1.getUtxoWallet)(coin, { id: '5b34252f1bf349930e34020a00000000', coin: coin.getChain() });
    const rootWalletKeys = (0, util_1.getDefaultWalletKeys)();
    const userPrv = rootWalletKeys.user.toBase58();
    const pubs = util_1.keychainsBase58.map((v) => v.pub);
    function validatePsbt(txHex, targetSigCount, targetNonceCount) {
        const psbt = utxolib.bitgo.createPsbtFromHex(txHex, coin.network);
        psbt.data.inputs.forEach((input, index) => {
            const parsed = utxolib.bitgo.parsePsbtInput(input);
            if (parsed.scriptType === 'taprootKeyPathSpend') {
                assert.ok(targetNonceCount);
                const nonce = psbt.getProprietaryKeyVals(index, {
                    identifier: utxolib.bitgo.PSBT_PROPRIETARY_IDENTIFIER,
                    subtype: utxolib.bitgo.ProprietaryKeySubtype.MUSIG2_PUB_NONCE,
                });
                assert.strictEqual(nonce.length, targetNonceCount);
            }
            const expectedSigCount = parsed.scriptType === 'p2shP2pk' || targetSigCount === 0 ? undefined : 1;
            assert.strictEqual(parsed.signatures?.length, expectedSigCount);
        });
    }
    function validateTx(txHex, unspents, targetSigCount) {
        const tx = utxolib.bitgo.createTransactionFromHex(txHex, coin.network);
        unspents.forEach((u, i) => {
            const sigCount = utxolib.bitgo.getStrictSignatureCount(tx.ins[i]);
            const expectedSigCount = utxolib.bitgo.isWalletUnspent(u) && !!targetSigCount ? 1 : 0;
            assert.strictEqual(sigCount, expectedSigCount);
        });
    }
    async function signTransaction(tx, useSigningSteps, unspents) {
        const isPsbt = tx instanceof utxolib.bitgo.UtxoPsbt;
        const isTxWithTaprootKeyPathSpend = isPsbt && utxolib.bitgo.isTransactionWithKeyPathSpendInput(tx);
        const txHex = tx.toHex();
        function nockSignPsbt(psbtHex) {
            const psbt = utxolib.bitgo.createPsbtFromHex(psbtHex, coin.network);
            return nock(bgUrl)
                .post(`/api/v2/${wallet.coin()}/wallet/${wallet.id()}/tx/signpsbt`, (body) => body.psbt)
                .reply(200, { psbt: psbt.setAllInputsMusig2NonceHD(rootWalletKeys.bitgo).toHex() });
        }
        if (!useSigningSteps) {
            let scope;
            if (tx instanceof utxolib.bitgo.UtxoPsbt && isTxWithTaprootKeyPathSpend) {
                scope = nockSignPsbt(tx.clone().setAllInputsMusig2NonceHD(rootWalletKeys.bitgo).toHex());
            }
            const psbt = await coin.signTransaction({
                txPrebuild: {
                    txHex,
                    txInfo: isPsbt ? undefined : { unspents },
                    walletId: isTxWithTaprootKeyPathSpend ? wallet.id() : undefined,
                },
                prv: userPrv,
                pubs: isPsbt ? undefined : pubs,
            });
            assert.ok('txHex' in psbt);
            if (isPsbt) {
                validatePsbt(psbt.txHex, 1, 2);
            }
            else {
                assert.ok(unspents);
                validateTx(psbt.txHex, unspents, 1);
            }
            if (scope) {
                assert.strictEqual(scope.isDone(), true);
            }
            return;
        }
        const signerNoncePsbt = await coin.signTransaction({
            txPrebuild: { txHex },
            prv: userPrv,
            signingStep: 'signerNonce',
        });
        assert.ok('txHex' in signerNoncePsbt);
        if (isPsbt) {
            validatePsbt(signerNoncePsbt.txHex, 0, isTxWithTaprootKeyPathSpend ? 1 : undefined);
        }
        else {
            assert.ok(unspents);
            validateTx(signerNoncePsbt.txHex, unspents, 0);
        }
        let scope;
        if (isTxWithTaprootKeyPathSpend) {
            scope = nockSignPsbt(signerNoncePsbt.txHex);
        }
        const cosignerNoncePsbt = await coin.signTransaction({
            txPrebuild: { ...signerNoncePsbt, walletId: wallet.id() },
            signingStep: 'cosignerNonce',
        });
        assert.ok('txHex' in cosignerNoncePsbt);
        if (isPsbt) {
            validatePsbt(cosignerNoncePsbt.txHex, 0, isTxWithTaprootKeyPathSpend ? 2 : undefined);
        }
        else {
            assert.ok(unspents);
            validateTx(cosignerNoncePsbt.txHex, unspents, 0);
        }
        if (scope) {
            assert.strictEqual(scope.isDone(), true);
        }
        const signerSigPsbt = await coin.signTransaction({
            txPrebuild: { ...cosignerNoncePsbt, txInfo: isPsbt ? undefined : { unspents } },
            prv: userPrv,
            pubs: isPsbt ? undefined : pubs,
            signingStep: 'signerSignature',
        });
        assert.ok('txHex' in signerSigPsbt);
        if (isPsbt) {
            validatePsbt(signerSigPsbt.txHex, 1, isTxWithTaprootKeyPathSpend ? 2 : undefined);
        }
        else {
            assert.ok(unspents);
            validateTx(signerSigPsbt.txHex, unspents, 1);
        }
    }
    it('customSigningFunction flow - PSBT with taprootKeyPathSpend inputs', async function () {
        const inputs = utxo_lib_1.testutil.inputScriptTypes.map((scriptType) => ({
            scriptType,
            value: BigInt(1000),
        }));
        const unspentSum = inputs.reduce((prev, curr) => prev + curr.value, BigInt(0));
        const outputs = [{ scriptType: 'p2sh', value: unspentSum - BigInt(1000) }];
        const psbt = utxo_lib_1.testutil.constructPsbt(inputs, outputs, coin.network, rootWalletKeys, 'unsigned', {
            p2shP2pkKey: (0, src_1.getReplayProtectionPubkeys)(coin.name)[0],
        });
        for (const v of [false, true]) {
            await signTransaction(psbt, v);
        }
    });
    it('customSigningFunction flow - PSBT without taprootKeyPathSpend inputs', async function () {
        const inputs = utxo_lib_1.testutil.inputScriptTypes
            .filter((v) => v !== 'taprootKeyPathSpend')
            .map((scriptType) => ({
            scriptType,
            value: BigInt(1000),
        }));
        const unspentSum = inputs.reduce((prev, cur) => prev + cur.value, BigInt(0));
        const outputs = [{ scriptType: 'p2sh', value: unspentSum - BigInt(1000) }];
        const psbt = utxo_lib_1.testutil.constructPsbt(inputs, outputs, coin.network, rootWalletKeys, 'unsigned');
        for (const v of [false, true]) {
            await signTransaction(psbt, v);
        }
    });
    it('customSigningFunction flow - Network Tx', async function () {
        const inputs = utxo_lib_1.testutil.txnInputScriptTypes
            .filter((v) => v !== 'p2shP2pk')
            .map((scriptType) => ({
            scriptType,
            value: BigInt(1000),
        }));
        const unspentSum = inputs.reduce((prev, curr) => prev + curr.value, BigInt(0));
        const outputs = [{ scriptType: 'p2sh', value: unspentSum - BigInt(1000) }];
        const txBuilder = utxo_lib_1.testutil.constructTxnBuilder(inputs, outputs, coin.network, rootWalletKeys, 'unsigned');
        const unspents = inputs.map((v, i) => utxo_lib_1.testutil.toTxnUnspent(v, i, coin.network, rootWalletKeys));
        for (const v of [false, true]) {
            await signTransaction(txBuilder.buildIncomplete(), v, unspents);
        }
    });
    it('fails on PSBT cache miss', async function () {
        const inputs = [{ scriptType: 'taprootKeyPathSpend', value: BigInt(1000) }];
        const unspentSum = inputs.reduce((prev, curr) => prev + curr.value, BigInt(0));
        const outputs = [{ scriptType: 'p2sh', value: unspentSum - BigInt(1000) }];
        const psbt = utxo_lib_1.testutil.constructPsbt(inputs, outputs, coin.network, rootWalletKeys, 'unsigned');
        await assert.rejects(async () => {
            await coin.signTransaction({
                txPrebuild: { txHex: psbt.toHex() },
                prv: userPrv,
                signingStep: 'signerSignature',
            });
        }, {
            message: `Psbt is missing from txCache (cache size 0).
            This may be due to the request being routed to a different BitGo-Express instance that for signing step 'signerNonce'.`,
        });
    });
    it('fails on unsupported locking script', async function () {
        const inputs = [
            { scriptType: 'p2wsh', value: BigInt(1000) },
            { scriptType: 'p2trMusig2', value: BigInt(1000) },
        ];
        const unspentSum = inputs.reduce((prev, curr) => prev + curr.value, BigInt(0));
        const outputs = [{ scriptType: 'p2sh', value: unspentSum - BigInt(500) }];
        const psbt = utxo_lib_1.testutil.constructPsbt(inputs, outputs, coin.network, rootWalletKeys, 'unsigned');
        // override the 1st PSBT input with unsupported 2 of 2 multi-sig locking script.
        const unspent = utxo_lib_1.testutil.toUnspent(inputs[0], 0, coin.network, rootWalletKeys);
        if (!utxolib.bitgo.isWalletUnspent(unspent)) {
            throw new Error('invalid unspent');
        }
        const { publicKeys } = rootWalletKeys.deriveForChainAndIndex(unspent.chain, unspent.index);
        const script2Of2 = utxolib.payments.p2ms({ m: 2, pubkeys: [publicKeys[0], publicKeys[1]] });
        psbt.data.inputs[0].witnessScript = script2Of2.output;
        await assert.rejects(async () => {
            await coin.signTransaction({
                txPrebuild: { txHex: psbt.toHex() },
                prv: userPrv,
            });
        }, {
            message: `length mismatch`,
        });
    });
});
