// @bitgo/abstract-utxo/dist/cjs/test/unit/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
require("mocha");
const assert = __importStar(require("assert"));
const _ = __importStar(require("lodash"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const nock = require("nock");
const utxo_lib_1 = require("@bitgo/utxo-lib");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../src");
const util_1 = require("./util");
function run(coin, inputScripts, txFormat, { decodeWith } = {}) {
    const amountType = coin.amountType;
    const title = [
        inputScripts.join(','),
        `txFormat=${txFormat}`,
        `amountType=${amountType}`,
        decodeWith ? `decodeWith=${decodeWith}` : '',
    ];
    describe(`${title.join(' ')}`, function () {
        const bgUrl = sdk_core_1.common.Environments[util_1.defaultBitGo.getEnv()].uri;
        const isTransactionWithKeyPathSpend = inputScripts.some((s) => s === 'taprootKeyPathSpend');
        const isTransactionWithReplayProtection = inputScripts.some((s) => s === 'p2shP2pk');
        const isTransactionWithP2tr = inputScripts.some((s) => s === 'p2tr');
        const isTransactionWithP2trMusig2 = inputScripts.some((s) => s === 'p2trMusig2');
        const value = (amountType === 'bigint' ? BigInt('10999999800000001') : 1e8);
        const wallet = (0, util_1.getUtxoWallet)(coin, { id: '5b34252f1bf349930e34020a00000000', coin: coin.getChain() });
        const walletKeys = (0, util_1.getDefaultWalletKeys)();
        const fullSign = !(isTransactionWithReplayProtection || isTransactionWithKeyPathSpend);
        function getUnspentsForPsbt() {
            return inputScripts.map((t, index) => {
                return utxo_lib_1.testutil.toUnspent({ scriptType: t, value: t === 'p2shP2pk' ? BigInt(1000) : BigInt(value) }, index, coin.network, walletKeys);
            });
        }
        function toTxnInputScriptType(type) {
            return type === 'p2shP2pk' ? 'replayProtection' : type === 'taprootKeyPathSpend' ? 'p2trMusig2' : type;
        }
        function getUnspents() {
            return inputScripts.map((type, i) => (0, util_1.mockUnspent)(coin.network, walletKeys, toTxnInputScriptType(type), i, value));
        }
        function getOutputAddress(rootWalletKeys) {
            return (0, src_1.generateAddress)(coin.name, {
                keychains: rootWalletKeys.triple.map((k) => ({ pub: k.neutered().toBase58() })),
            });
        }
        function getSignParams(prebuildHex, signer, cosigner, decodeWith) {
            const txInfo = {
                unspents: txFormat === 'psbt' ? undefined : getUnspents(),
            };
            return {
                txPrebuild: {
                    walletId: isTransactionWithKeyPathSpend ? wallet.id() : undefined,
                    txHex: prebuildHex,
                    txInfo,
                    decodeWith,
                },
                prv: signer.toBase58(),
                pubs: walletKeys.triple.map((k) => k.neutered().toBase58()),
                cosignerPub: cosigner.neutered().toBase58(),
            };
        }
        async function createHalfSignedTransaction(prebuild, signer, cosigner) {
            let scope;
            if (prebuild instanceof utxolib.bitgo.UtxoPsbt && isTransactionWithKeyPathSpend) {
                const psbt = prebuild.clone().setAllInputsMusig2NonceHD(cosigner);
                scope = nock(bgUrl)
                    .post(`/api/v2/${wallet.coin()}/wallet/${wallet.id()}/tx/signpsbt`, (body) => body.psbt)
                    .reply(200, { psbt: psbt.toHex() });
            }
            // half-sign with the user key
            const result = (await wallet.signTransaction(getSignParams(prebuild.toBuffer().toString('hex'), signer, cosigner, decodeWith)));
            if (scope) {
                assert.strictEqual(scope.isDone(), true);
            }
            return result;
        }
        async function createFullSignedTransaction(halfSigned, signer, cosigner) {
            return (await wallet.signTransaction({
                ...getSignParams(halfSigned.txHex, signer, cosigner, decodeWith),
                isLastSignature: true,
            }));
        }
        function createPrebuildPsbt() {
            const inputs = inputScripts.map((t) => ({
                scriptType: t,
                value: t === 'p2shP2pk' ? BigInt(1000) : BigInt(value),
            }));
            const unspentSum = inputs.reduce((prev, curr) => prev + curr.value, BigInt(0));
            const outputs = [
                { address: getOutputAddress((0, util_1.getWalletKeys)('test')), value: unspentSum - BigInt(1000) },
            ];
            const psbt = utxo_lib_1.testutil.constructPsbt(inputs, outputs, coin.network, walletKeys, 'unsigned', {
                p2shP2pkKey: (0, src_1.getReplayProtectionPubkeys)(coin.name)[0],
            });
            utxolib.bitgo.addXpubsToPsbt(psbt, walletKeys);
            return psbt;
        }
        async function getTransactionStages() {
            const prebuild = txFormat === 'psbt'
                ? createPrebuildPsbt()
                : (0, util_1.createPrebuildTransaction)(coin.network, getUnspents(), getOutputAddress(walletKeys));
            const halfSignedUserBitGo = await createHalfSignedTransaction(prebuild, walletKeys.user, walletKeys.bitgo);
            const fullSignedUserBitGo = fullSign && !isTransactionWithP2trMusig2
                ? await createFullSignedTransaction(halfSignedUserBitGo, walletKeys.bitgo, walletKeys.user)
                : undefined;
            const halfSignedUserBackup = !isTransactionWithKeyPathSpend && !(txFormat === 'psbt' && isTransactionWithP2tr)
                ? await createHalfSignedTransaction(prebuild, walletKeys.user, walletKeys.backup)
                : undefined;
            const fullSignedUserBackup = fullSign && halfSignedUserBackup
                ? await createFullSignedTransaction(halfSignedUserBackup, walletKeys.backup, walletKeys.user)
                : undefined;
            return {
                prebuild,
                halfSignedUserBackup,
                halfSignedUserBitGo,
                fullSignedUserBackup,
                fullSignedUserBitGo,
            };
        }
        let transactionStages;
        before('prepare', async function () {
            transactionStages = await getTransactionStages();
        });
        afterEach(nock.cleanAll);
        it('match fixtures', async function () {
            if (txFormat === 'psbt') {
                // TODO (maybe) - once full PSBT support is added to abstract-utxo module, custom JSON representation of PSBT can be created and tested here.
                // signatures of taprootKeyPathSpends are random since random nature of MuSig2 nonce, so psbt hex comparison also wont work.
                return this.skip();
            }
            function toTransactionStagesObj(stages) {
                return _.mapValues(stages, (v) => v === undefined || v instanceof utxolib.bitgo.UtxoPsbt
                    ? undefined
                    : v instanceof utxolib.bitgo.UtxoTransaction
                        ? (0, util_1.transactionToObj)(v)
                        : (0, util_1.transactionHexToObj)(v.txHex, coin.network, amountType));
            }
            (0, util_1.shouldEqualJSON)(toTransactionStagesObj(transactionStages), await (0, util_1.getFixture)(coin, `transactions-${inputScripts.map((t) => toTxnInputScriptType(t)).join('-')}`, toTransactionStagesObj(transactionStages)));
        });
        function testPsbtValidSignatures(tx, signedBy) {
            const psbt = utxolib.bitgo.createPsbtFromHex(tx.txHex, coin.network);
            const unspents = getUnspentsForPsbt();
            psbt.data.inputs.forEach((input, index) => {
                const unspent = unspents[index];
                if (!utxolib.bitgo.isWalletUnspent(unspent)) {
                    assert.ok(utxolib.bitgo.getPsbtInputScriptType(input), 'p2shP2pk');
                    return;
                }
                const walletUnspent = unspent;
                const pubkeys = walletKeys.deriveForChainAndIndex(walletUnspent.chain, walletUnspent.index).publicKeys;
                pubkeys.forEach((pk, pkIndex) => {
                    psbt.validateSignaturesOfInputCommon(index, pk).should.eql(signedBy.includes(walletKeys.triple[pkIndex]));
                });
            });
        }
        function testValidSignatures(tx, signedBy, sign) {
            if (txFormat === 'psbt' && sign === 'halfsigned') {
                testPsbtValidSignatures(tx, signedBy);
                return;
            }
            const unspents = txFormat === 'psbt'
                ? getUnspentsForPsbt().map((u) => ({ ...u, value: utxo_lib_1.bitgo.toTNumber(u.value, amountType) }))
                : getUnspents();
            const prevOutputs = unspents.map((u) => ({
                script: Buffer.from(wasm_utxo_1.address.toOutputScriptWithCoin(u.address, coin.name)),
                value: u.value,
            }));
            const transaction = utxolib.bitgo.createTransactionFromBuffer(Buffer.from(tx.txHex, 'hex'), coin.network, { amountType });
            transaction.ins.forEach((input, index) => {
                if (inputScripts[index] === 'p2shP2pk') {
                    assert.ok(coin.isBitGoTaintedUnspent(unspents[index]));
                    return;
                }
                const unspent = unspents[index];
                const pubkeys = walletKeys.deriveForChainAndIndex(unspent.chain, unspent.index).publicKeys;
                pubkeys.forEach((pk, pkIndex) => {
                    utxolib.bitgo
                        .verifySignature(transaction, index, prevOutputs[index].value, {
                        publicKey: pk,
                    }, prevOutputs)
                        .should.eql(signedBy.includes(walletKeys.triple[pkIndex]));
                });
            });
        }
        async function testExplainTx(stageName, txHex, unspents, pubs) {
            const explanation = await coin.explainTransaction({
                txHex,
                txInfo: {
                    unspents,
                },
                pubs,
                decodeWith,
            });
            const expectedProperties = ['id', 'outputs', 'changeOutputs', 'changeAmount', 'outputAmount'];
            if (decodeWith !== 'wasm-utxo') {
                expectedProperties.push('displayOrder', 'inputSignatures', 'signatures');
            }
            explanation.should.have.properties(...expectedProperties);
            const expectedSignatureCount = stageName === 'prebuild' || pubs === undefined
                ? 0
                : stageName.startsWith('halfSigned')
                    ? 1
                    : stageName.startsWith('fullSigned')
                        ? 2
                        : undefined;
            if ('inputSignatures' in explanation) {
                explanation.inputSignatures.should.eql(
                // FIXME(BG-35154): implement signature verification for replay protection inputs
                inputScripts.map((type) => (type === 'p2shP2pk' ? 0 : expectedSignatureCount)));
            }
            if ('signatures' in explanation) {
                explanation.signatures.should.eql(expectedSignatureCount);
                explanation.changeAmount.should.eql('0'); // no change addresses given
            }
            let expectedOutputAmount = BigInt((txFormat === 'psbt' ? getUnspentsForPsbt() : getUnspents()).length) * BigInt(value);
            inputScripts.forEach((type) => {
                if (type === 'p2shP2pk') {
                    // replayProtection unspents have value 1000
                    expectedOutputAmount -= BigInt(value);
                    expectedOutputAmount += BigInt(1000);
                }
            });
            expectedOutputAmount -= BigInt(1000); // fee of 1000
            explanation.outputAmount.should.eql(expectedOutputAmount.toString());
        }
        it('have valid signature for half-signed transaction', function () {
            if (transactionStages.halfSignedUserBackup) {
                testValidSignatures(transactionStages.halfSignedUserBackup, [walletKeys.user], 'halfsigned');
            }
            testValidSignatures(transactionStages.halfSignedUserBitGo, [walletKeys.user], 'halfsigned');
        });
        it('have valid signatures for full-signed transaction', function () {
            if (!fullSign) {
                return this.skip();
            }
            if (transactionStages.fullSignedUserBackup) {
                testValidSignatures(transactionStages.fullSignedUserBackup, [walletKeys.user, walletKeys.backup], 'fullsigned');
            }
            if (transactionStages.fullSignedUserBitGo) {
                testValidSignatures(transactionStages.fullSignedUserBitGo, [walletKeys.user, walletKeys.bitgo], 'fullsigned');
            }
        });
        it('have correct results for explainTransaction', async function () {
            for (const [stageName, stageTx] of Object.entries(transactionStages)) {
                if (!stageTx) {
                    continue;
                }
                const txHex = stageTx instanceof utxolib.bitgo.UtxoPsbt || stageTx instanceof utxolib.bitgo.UtxoTransaction
                    ? stageTx.toBuffer().toString('hex')
                    : stageTx.txHex;
                const pubs = walletKeys.triple.map((k) => k.neutered().toBase58());
                const unspents = txFormat === 'psbt'
                    ? getUnspentsForPsbt().map((u) => ({ ...u, value: utxo_lib_1.bitgo.toTNumber(u.value, amountType) }))
                    : getUnspents();
                await testExplainTx(stageName, txHex, unspents, pubs);
                if (decodeWith !== 'wasm-utxo') {
                    await testExplainTx(stageName, txHex, unspents, undefined);
                }
            }
        });
    });
}
function runTestForCoin(coin) {
    ['legacy', 'psbt'].forEach((txFormat) => {
        run(coin, (0, util_1.getScriptTypes)(coin, txFormat), txFormat, { decodeWith: 'wasm-utxo' });
    });
}
describe('Transaction Suite', function () {
    (0, util_1.getMinUtxoCoins)().forEach((coin) => {
        describe(`${coin.getChain()}`, function () {
            runTestForCoin(coin);
        });
    });
});
