// @bitgo/abstract-utxo/dist/cjs/test/unit/txFormat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWalletConfigurations = getWalletConfigurations;
exports.createMockWallet = createMockWallet;
exports.getTxFormat = getTxFormat;
const assert = __importStar(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../src");
const names_1 = require("../../src/names");
const util_1 = require("./util");
/**
 * Enumerates common wallet configurations for testing
 */
function getWalletConfigurations() {
    return [
        { name: 'hot wallet', options: { type: 'hot' } },
        { name: 'cold wallet', options: { type: 'cold' } },
        { name: 'custodial wallet', options: { type: 'custodial' } },
        { name: 'distributedCustody wallet', options: { type: 'cold', subType: 'distributedCustody' } },
        { name: 'musigKp wallet', options: { type: 'cold', walletFlags: [{ name: 'musigKp', value: 'true' }] } },
        { name: 'hot musigKp wallet', options: { type: 'hot', walletFlags: [{ name: 'musigKp', value: 'true' }] } },
    ];
}
/**
 * Helper function to create a mock wallet for testing
 */
function createMockWallet(coin, options = {}) {
    const walletData = {
        id: '5b34252f1bf349930e34020a',
        coin: coin.getChain(),
        type: options.type || 'hot',
        ...(options.subType && { subType: options.subType }),
        ...(options.walletFlags && { walletFlags: options.walletFlags }),
    };
    return new sdk_core_1.Wallet(util_1.defaultBitGo, coin, walletData);
}
/**
 * Helper function to get the txFormat from a coin's getDefaultTxFormat method
 */
function getTxFormat(coin, wallet, requestedFormat) {
    return coin.getDefaultTxFormat(wallet, requestedFormat);
}
/**
 * Helper function to run a txFormat test with named arguments.
 * By default, iterates over all wallet configurations and all coins.
 */
function runTest(params) {
    it(params.description, function () {
        const walletConfigs = getWalletConfigurations();
        for (const walletConfig of walletConfigs) {
            // Skip wallet configurations that don't match the filter
            if (params.walletFilter && !params.walletFilter(walletConfig)) {
                continue;
            }
            for (const coin of util_1.utxoCoins) {
                // Skip coins that don't match the filter
                if (params.coinFilter && !params.coinFilter(coin)) {
                    continue;
                }
                const wallet = createMockWallet(coin, walletConfig.options);
                const txFormat = getTxFormat(coin, wallet, params.requestedTxFormat);
                const expectedTxFormat = typeof params.expectedTxFormat === 'function'
                    ? params.expectedTxFormat(coin, walletConfig.options)
                    : params.expectedTxFormat;
                assert.strictEqual(txFormat, expectedTxFormat, `${params.description} - ${walletConfig.name} - ${coin.getChain()}: expected ${expectedTxFormat}, got ${txFormat}`);
            }
        }
    });
}
describe('txFormat', function () {
    describe('getDefaultTxFormat', function () {
        // All testnet wallets default to PSBT-lite
        runTest({
            description: 'should always return psbt-lite for testnet',
            coinFilter: (coin) => (0, names_1.isTestnetCoin)(coin.name),
            expectedTxFormat: 'psbt-lite',
        });
        // All mainnet wallets default to psbt-lite
        runTest({
            description: 'should return psbt-lite for all mainnet wallets',
            coinFilter: (coin) => (0, names_1.isMainnetCoin)(coin.name),
            expectedTxFormat: 'psbt-lite',
        });
        // Test explicitly requested formats
        runTest({
            description: 'should respect explicitly requested legacy format on mainnet',
            coinFilter: (coin) => (0, names_1.isMainnetCoin)(coin.name),
            expectedTxFormat: 'legacy',
            requestedTxFormat: 'legacy',
        });
        runTest({
            description: 'should respect explicitly requested psbt format',
            expectedTxFormat: 'psbt',
            requestedTxFormat: 'psbt',
        });
        runTest({
            description: 'should respect explicitly requested psbt-lite format',
            expectedTxFormat: 'psbt-lite',
            requestedTxFormat: 'psbt-lite',
        });
        // Test that legacy format is prohibited on testnet
        it('should throw ErrorDeprecatedTxFormat when legacy format is requested on testnet', function () {
            for (const coin of util_1.utxoCoins) {
                if (!(0, names_1.isTestnetCoin)(coin.name)) {
                    continue;
                }
                const wallet = createMockWallet(coin, { type: 'hot' });
                assert.throws(() => getTxFormat(coin, wallet, 'legacy'), src_1.ErrorDeprecatedTxFormat, `Expected ErrorDeprecatedTxFormat for ${coin.getChain()}`);
            }
        });
    });
});
