// @bitgo/abstract-utxo/dist/cjs/test/unit/verifyTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const sinon = __importStar(require("sinon"));
const sdk_core_1 = require("@bitgo/sdk-core");
const util_1 = require("./util");
describe('Verify Transaction', function () {
    const coin = (0, util_1.getUtxoCoin)('tbtc');
    const userKeychain = coin.keychains().create();
    const otherKeychain = coin.keychains().create();
    const changeKeys = {
        user: coin.keychains().create(),
        backup: coin.keychains().create(),
        bitgo: coin.keychains().create(),
    };
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const sign = async (key, keychain) => (await coin.signMessage(keychain, key.pub)).toString('hex');
    const signUser = (key) => sign(key, userKeychain);
    const signOther = (key) => sign(key, otherKeychain);
    const passphrase = 'test_passphrase';
    const stubData = {
        unsignedSendingWallet: {
            keyIds: sinon.stub().returns(['0', '1', '2']),
        },
        parseTransactionData: {
            badKey: {
                keychains: {
                    // user public key swapped out
                    user: {
                        pub: otherKeychain.pub,
                        encryptedPrv: util_1.defaultBitGo.encrypt({
                            input: userKeychain.prv,
                            password: passphrase,
                        }),
                    },
                },
                needsCustomChangeKeySignatureVerification: true,
            },
            noCustomChange: {
                keychains: { user: userKeychain },
                needsCustomChangeKeySignatureVerification: true,
            },
            emptyCustomChange: {
                keychains: { user: userKeychain },
                needsCustomChangeKeySignatureVerification: true,
                customChange: {},
            },
            // needs to be async function to create signatures
            badSigs: async () => ({
                keychains: { user: userKeychain },
                needsCustomChangeKeySignatureVerification: true,
                customChange: {
                    keys: [changeKeys.user, changeKeys.backup, changeKeys.bitgo],
                    signatures: [
                        await signOther(changeKeys.user),
                        await signOther(changeKeys.backup),
                        await signOther(changeKeys.bitgo),
                    ],
                },
            }),
            goodSigs: async () => ({
                keychains: { user: userKeychain },
                needsCustomChangeKeySignatureVerification: true,
                customChange: {
                    keys: [changeKeys.user, changeKeys.backup, changeKeys.bitgo],
                    signatures: [
                        await signUser(changeKeys.user),
                        await signUser(changeKeys.backup),
                        await signUser(changeKeys.bitgo),
                    ],
                },
                missingOutputs: 1,
            }),
        },
    };
    const unsignedSendingWallet = sinon.createStubInstance(sdk_core_1.Wallet, stubData.unsignedSendingWallet);
    it('should fail if the user private key cannot be verified to match the user public key', async () => {
        sinon.stub(coin, 'parseTransaction').resolves(stubData.parseTransactionData.badKey);
        const verifyWallet = sinon.createStubInstance(sdk_core_1.Wallet, {});
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {},
            wallet: verifyWallet,
            verification: {},
        }), /transaction requires verification of user public key, but it was unable to be verified/);
        coin.parseTransaction.restore();
    });
    it('should fail if the custom change verification data is required but missing', async () => {
        sinon.stub(coin, 'parseTransaction').resolves(stubData.parseTransactionData.noCustomChange);
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {},
            wallet: unsignedSendingWallet,
            verification: {},
        }), /parsed transaction is missing required custom change verification data/);
        coin.parseTransaction.restore();
    });
    it('should fail if the custom change keys or key signatures are missing', async () => {
        sinon.stub(coin, 'parseTransaction').resolves(stubData.parseTransactionData.emptyCustomChange);
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {},
            wallet: unsignedSendingWallet,
            verification: {},
        }), /customChange property is missing keys or signatures/);
        coin.parseTransaction.restore();
    });
    it('should fail if the custom change key signatures cannot be verified', async () => {
        sinon.stub(coin, 'parseTransaction').resolves((await stubData.parseTransactionData.badSigs()));
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {},
            wallet: unsignedSendingWallet,
            verification: {},
        }), /transaction requires verification of custom change key signatures, but they were unable to be verified/);
        coin.parseTransaction.restore();
    });
    it('should successfully verify a custom change transaction when change keys and signatures are valid', async () => {
        sinon.stub(coin, 'parseTransaction').resolves((await stubData.parseTransactionData.goodSigs()));
        // if verify transaction gets rejected with the outputs missing error message,
        // then we know that the verification of the custom change key signatures was successful
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {},
            wallet: unsignedSendingWallet,
            verification: {},
        }), /expected outputs missing in transaction prebuild/);
        coin.parseTransaction.restore();
    });
    it('should not allow more than 150 basis points of implicit external outputs (for paygo outputs)', async () => {
        const coinMock = sinon.stub(coin, 'parseTransaction').resolves({
            keychains: {},
            keySignatures: {},
            outputs: [],
            missingOutputs: [],
            explicitExternalOutputs: [],
            implicitExternalOutputs: [],
            changeOutputs: [],
            explicitExternalSpendAmount: 10000,
            implicitExternalSpendAmount: 151,
            needsCustomChangeKeySignatureVerification: false,
        });
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {},
            wallet: unsignedSendingWallet,
        }), /prebuild attempts to spend to unintended external recipients/);
        coinMock.restore();
    });
    it('should allow 150 basis points of implicit external outputs (for paygo outputs)', async () => {
        const coinMock = sinon.stub(coin, 'parseTransaction').resolves({
            keychains: {},
            keySignatures: {},
            outputs: [],
            missingOutputs: [],
            explicitExternalOutputs: [],
            implicitExternalOutputs: [],
            changeOutputs: [],
            explicitExternalSpendAmount: 1000,
            implicitExternalSpendAmount: 15,
            needsCustomChangeKeySignatureVerification: false,
        });
        const bitcoinMock = sinon
            .stub(coin, 'createTransactionFromHex')
            .returns({ ins: [] });
        const result = await coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {
                txHex: '00',
            },
            wallet: unsignedSendingWallet,
        });
        assert_1.default.strictEqual(result, true);
        coinMock.restore();
        bitcoinMock.restore();
    });
    it('should not allow any implicit external outputs if paygo outputs are disallowed', async () => {
        const coinMock = sinon.stub(coin, 'parseTransaction').resolves({
            keychains: {},
            keySignatures: {},
            outputs: [],
            missingOutputs: [],
            explicitExternalOutputs: [],
            implicitExternalOutputs: [],
            changeOutputs: [],
            explicitExternalSpendAmount: 0,
            implicitExternalSpendAmount: 10,
            needsCustomChangeKeySignatureVerification: false,
        });
        await assert_1.default.rejects(coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {
                txHex: '00',
            },
            wallet: unsignedSendingWallet,
            verification: {
                allowPaygoOutput: false,
            },
        }), /prebuild attempts to spend to unintended external recipients/);
        coinMock.restore();
    });
    it('should allow paygo outputs if empty verification object is passed', async () => {
        const coinMock = sinon.stub(coin, 'parseTransaction').resolves({
            keychains: {},
            keySignatures: {},
            outputs: [],
            missingOutputs: [],
            explicitExternalOutputs: [],
            implicitExternalOutputs: [],
            changeOutputs: [],
            explicitExternalSpendAmount: 1000,
            implicitExternalSpendAmount: 15,
            needsCustomChangeKeySignatureVerification: false,
        });
        const bitcoinMock = sinon
            .stub(coin, 'createTransactionFromHex')
            .returns({ ins: [] });
        const result = await coin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {
                txHex: '00',
            },
            wallet: unsignedSendingWallet,
            verification: {},
        });
        assert_1.default.strictEqual(result, true);
        coinMock.restore();
        bitcoinMock.restore();
    });
    it('should work with bigint amounts', async () => {
        // need a coin that uses bigint
        const bigintCoin = (0, util_1.getUtxoCoin)('tdoge');
        const coinMock = sinon.stub(bigintCoin, 'parseTransaction').resolves({
            keychains: {},
            keySignatures: {},
            outputs: [],
            missingOutputs: [],
            explicitExternalOutputs: [
                {
                    address: 'external_address',
                    amount: '10000',
                },
            ],
            implicitExternalOutputs: [
                {
                    address: 'external_address_2',
                    amount: '15',
                },
            ],
            changeOutputs: [],
            explicitExternalSpendAmount: BigInt(10000),
            implicitExternalSpendAmount: BigInt(15),
            needsCustomChangeKeySignatureVerification: false,
        });
        const bitcoinMock = sinon
            .stub(bigintCoin, 'createTransactionFromHex')
            .returns({ ins: [] });
        const result = await bigintCoin.verifyTransaction({
            txParams: {
                walletPassphrase: passphrase,
            },
            txPrebuild: {
                txHex: '00',
            },
            wallet: unsignedSendingWallet,
            verification: {},
        });
        assert_1.default.strictEqual(result, true);
        coinMock.restore();
        bitcoinMock.restore();
    });
});
