// @bitgo/sdk-coin-xdc/dist/src/lib/transactionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const utils_1 = require("./utils");
const transferBuilder_1 = require("./transferBuilder");
const uploadKycBuilder_1 = require("./uploadKycBuilder");
class TransactionBuilder extends abstract_eth_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._common = (0, utils_1.getCommon)(this._coinConfig.network.type);
        this.transaction = new abstract_eth_1.Transaction(this._coinConfig, this._common);
        this._signatures = [];
    }
    /** @inheritdoc */
    transfer(data) {
        if (this._type !== sdk_core_1.TransactionType.Send) {
            throw new sdk_core_1.BuildTransactionError('Transfers can only be set for send transactions');
        }
        if (!this._transfer) {
            this._transfer = new transferBuilder_1.TransferBuilder(data);
        }
        return this._transfer;
    }
    /**
     * Gets the uploadKYC builder if it exists, or creates a new one for this transaction
     *
     * @returns {UploadKycBuilder} the uploadKYC builder
     * @throws {BuildTransactionError} if transaction type is not ContractCall
     */
    uploadKyc() {
        if (this._type !== sdk_core_1.TransactionType.ContractCall) {
            throw new sdk_core_1.BuildTransactionError('uploadKYC can only be set for contract call transactions');
        }
        if (!this._uploadKycBuilder) {
            this._uploadKycBuilder = new uploadKycBuilder_1.UploadKycBuilder(this._coinConfig);
        }
        return this._uploadKycBuilder;
    }
    /** @inheritdoc */
    async build() {
        // If uploadKYC builder is set, build and cache the result before validation
        if (this._uploadKycBuilder && this._type === sdk_core_1.TransactionType.ContractCall) {
            this._uploadKycCall = this._uploadKycBuilder.build();
            this.contract(this._uploadKycCall.contractAddress);
            this.data(this._uploadKycCall.serialize());
        }
        return (await super.build());
    }
    /**
     * Build the transaction data for uploadKYC contract call
     * @private
     */
    buildUploadKycTransaction() {
        if (!this._uploadKycCall) {
            throw new sdk_core_1.BuildTransactionError('uploadKYC call not initialized');
        }
        // The contract address and data are already set in build() method
        // Just return the base transaction data with the serialized data
        return this.buildBase(this._uploadKycCall.serialize());
    }
    /** @inheritdoc */
    getTransactionData() {
        // If uploadKYC builder is set, build uploadKYC transaction
        if (this._uploadKycBuilder && this._type === sdk_core_1.TransactionType.ContractCall) {
            return this.buildUploadKycTransaction();
        }
        // Otherwise, use the parent implementation
        return super.getTransactionData();
    }
    addSignature(publicKey, signature) {
        this._signatures = [];
        this._signatures.push({ publicKey, signature });
    }
    getContractData(addresses) {
        throw new Error('Method not implemented.');
    }
}
exports.TransactionBuilder = TransactionBuilder;
