// @bitgo/sdk-coin-xdc/dist/src/lib/uploadKycBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UploadKycBuilder = exports.UploadKycCall = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const validatorContract_1 = require("./validatorContract");
/**
 * Represents an XDC uploadKYC contract call
 * This is used to submit KYC document hashes (IPFS hashes) to the XDC Validator contract
 */
class UploadKycCall extends abstract_eth_1.ContractCall {
    constructor(contractAddress, ipfsHash) {
        // uploadKYC(string) - takes a single string parameter (the IPFS hash)
        super(validatorContract_1.UPLOAD_KYC_METHOD_ID, ['string'], [ipfsHash]);
        this.contractAddress = contractAddress;
        this.ipfsHash = ipfsHash;
    }
}
exports.UploadKycCall = UploadKycCall;
/**
 * Builder for XDC uploadKYC transactions
 *
 * This builder creates transactions that upload KYC document hashes to the XDC Validator contract.
 * The flow is:
 * 1. Upload KYC document to IPFS → get IPFS hash (e.g., "Qm...")
 * 2. Use this builder to create a transaction that submits the hash to the validator contract
 * 3. The transaction is signed using TSS/MPC
 * 4. After successful upload, the address can propose a validator candidate
 *
 * @example
 * ```typescript
 * const builder = new UploadKycBuilder(coinConfig);
 * const call = builder
 *   .ipfsHash('QmRealIPFSHash...')
 *   .build();
 * ```
 */
class UploadKycBuilder {
    constructor(coinConfig) {
        this._coinConfig = coinConfig;
    }
    /**
     * Set the IPFS hash of the uploaded KYC document
     *
     * @param {string} hash - The IPFS hash (e.g., "QmRealIPFSHash...")
     * @returns {UploadKycBuilder} this builder instance
     * @throws {InvalidParameterValueError} if the hash is invalid
     */
    ipfsHash(hash) {
        if (!hash || hash.trim().length === 0) {
            throw new sdk_core_1.InvalidParameterValueError('IPFS hash cannot be empty');
        }
        // Basic IPFS hash validation (should start with 'Qm' for v0 or 'b' for v1)
        if (!hash.startsWith('Qm') && !hash.startsWith('b')) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid IPFS hash format. Expected hash starting with "Qm" (v0) or "b" (v1)');
        }
        this._ipfsHash = hash;
        return this;
    }
    /**
     * Set a custom validator contract address
     * If not set, the default address for the network will be used
     *
     * @param {string} address - The validator contract address
     * @returns {UploadKycBuilder} this builder instance
     * @throws {InvalidParameterValueError} if the address is invalid
     */
    contractAddress(address) {
        if (!address || address.trim().length === 0) {
            throw new sdk_core_1.InvalidParameterValueError('Contract address cannot be empty');
        }
        // Basic Ethereum address validation
        if (!/^(0x)?[0-9a-fA-F]{40}$/.test(address)) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid contract address format');
        }
        this._contractAddress = address.toLowerCase().startsWith('0x') ? address : `0x${address}`;
        return this;
    }
    /**
     * Build the uploadKYC contract call
     *
     * @returns {UploadKycCall} the constructed contract call
     * @throws {BuildTransactionError} if required fields are missing
     */
    build() {
        this.validateMandatoryFields();
        const contractAddress = this._contractAddress || this.getDefaultContractAddress();
        const ipfsHash = this._ipfsHash; // validated by validateMandatoryFields
        return new UploadKycCall(contractAddress, ipfsHash);
    }
    /**
     * Validate that all mandatory fields are set
     * @private
     */
    validateMandatoryFields() {
        if (!this._ipfsHash) {
            throw new sdk_core_1.BuildTransactionError('Missing IPFS hash for uploadKYC transaction');
        }
    }
    /**
     * Get the default validator contract address for the current network
     * @private
     */
    getDefaultContractAddress() {
        const isTestnet = this._coinConfig.name === 'txdc';
        return (0, validatorContract_1.getValidatorContractAddress)(isTestnet);
    }
}
exports.UploadKycBuilder = UploadKycBuilder;
