// @bitgo/sdk-coin-xdc/dist/src/lib/validatorContract.js
"use strict";
/**
 * XDC Validator Contract Constants and ABI
 *
 * This file contains the contract address and ABI for the XDC Validator contract
 * which handles KYC uploads and validator staking operations.
 *
 * Contract Address: 0x0000000000000000000000000000000000000088 (System Contract)
 * Source: XDC Network Validator Contract
 * Reference: https://github.com/XinFinOrg/XDPoSChain
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPOSE_ABI = exports.UPLOAD_KYC_ABI = exports.XDC_VALIDATOR_ABI = exports.PROPOSE_METHOD_ID = exports.UPLOAD_KYC_METHOD_ID = exports.XDC_VALIDATOR_CONTRACT_ADDRESS_MAINNET = exports.XDC_VALIDATOR_CONTRACT_ADDRESS_TESTNET = void 0;
exports.getValidatorContractAddress = getValidatorContractAddress;
const XDCValidatorABI_json_1 = __importDefault(require("./XDCValidatorABI.json"));
/**
 * XDC Validator Contract Address (Testnet)
 * This is the deployed contract address on XDC Apothem testnet (chainId: 51)
 */
exports.XDC_VALIDATOR_CONTRACT_ADDRESS_TESTNET = '0x0000000000000000000000000000000000000088';
/**
 * XDC Validator Contract Address (Mainnet)
 * This is the deployed contract address on XDC mainnet (chainId: 50)
 */
exports.XDC_VALIDATOR_CONTRACT_ADDRESS_MAINNET = '0x0000000000000000000000000000000000000088';
/**
 * uploadKYC method ID
 * keccak256("uploadKYC(string)") = 0xf5c95125
 */
exports.UPLOAD_KYC_METHOD_ID = '0xf5c95125';
/**
 * propose method ID
 * keccak256("propose(address)") = 0x01267951
 */
exports.PROPOSE_METHOD_ID = '0x01267951';
/**
 * Full XDC Validator Contract ABI
 * Imported from JSON file for easy use with Web3/XDC3
 */
exports.XDC_VALIDATOR_ABI = XDCValidatorABI_json_1.default;
/**
 * Get the uploadKYC function ABI
 * Useful for encoding function calls
 */
exports.UPLOAD_KYC_ABI = XDCValidatorABI_json_1.default.find((item) => item.name === 'uploadKYC');
/**
 * Get the propose function ABI
 * Useful for encoding function calls
 */
exports.PROPOSE_ABI = XDCValidatorABI_json_1.default.find((item) => item.name === 'propose');
/**
 * Get the validator contract address for the given network
 * @param isTestnet - whether to use testnet or mainnet
 * @returns the validator contract address
 */
function getValidatorContractAddress(isTestnet) {
    return isTestnet ? exports.XDC_VALIDATOR_CONTRACT_ADDRESS_TESTNET : exports.XDC_VALIDATOR_CONTRACT_ADDRESS_MAINNET;
}
