// @bitgo/sdk-coin-hbar/dist/src/lib/tokenTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const long_1 = __importDefault(require("long"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transferBuilder_1 = require("./transferBuilder");
const utils_1 = require("./utils");
const bignumber_js_1 = require("bignumber.js");
class TokenTransferBuilder extends transferBuilder_1.TransferBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    async buildImplementation() {
        this._txBodyData.tokenTransfers = this.buildTokenTransferData(); // set to list by the contract
        return await super.buildImplementation();
    }
    buildTokenTransferData() {
        let tokenTransferAmount = new bignumber_js_1.BigNumber(0); // total send amount for each token
        const tokenId = (0, utils_1.getHederaTokenIdFromName)(this._tokenName);
        const tokenTransferData = [
            {
                accountID: (0, utils_1.buildHederaAccountID)(this._source.address),
                amount: long_1.default.fromInt(0),
            },
        ];
        this._recipients.forEach((recipient) => {
            tokenTransferAmount = tokenTransferAmount.plus(recipient.amount);
            tokenTransferData.push({ accountID: (0, utils_1.buildHederaAccountID)(recipient.address), amount: long_1.default.fromString(recipient.amount) } // recipient
            );
        });
        tokenTransferData[0].amount = long_1.default.fromString(tokenTransferAmount.toString()).negate(); // update sender send amount
        return [
            {
                token: (0, utils_1.buildHederaTokenID)(tokenId),
                transfers: tokenTransferData,
            },
        ];
    }
    /** @inheritdoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        const transferData = tx.txBody.cryptoTransfer;
        if ((0, utils_1.isTokenTransfer)(transferData)) {
            this.initTokenTransfers(transferData.tokenTransfers);
        }
    }
    /**
     * Initialize the transfer specific data, getting the recipient account
     * represented by the element with a positive amount on the transfer element.
     * The negative amount represents the source account so it's ignored.
     *
     * @param {proto.IAccountAmount[]} transfers array of objects which contains accountID and transferred amount
     */
    initTokenTransfers(tokenTransfers) {
        tokenTransfers.forEach((tokenTransfer) => {
            if (!tokenTransfer.token) {
                throw new sdk_core_1.BuildTransactionError('Invalid transaction: missing token id');
            }
            if (!tokenTransfer.transfers) {
                throw new sdk_core_1.BuildTransactionError('Invalid transaction: missing transfer data');
            }
            const token = (0, utils_1.getHederaTokenNameFromId)((0, utils_1.stringifyTokenId)(tokenTransfer.token));
            if (!token) {
                throw new sdk_core_1.BuildTransactionError('Invalid transaction: invalid token id');
            }
            tokenTransfer.transfers.forEach((transferData) => {
                const amount = long_1.default.fromValue(transferData.amount);
                if (amount.isPositive()) {
                    this.send({
                        address: (0, utils_1.stringifyAccountId)(transferData.accountID),
                        amount: amount.toString(),
                        tokenName: token.name,
                    });
                }
            });
        });
    }
    // region Transfer fields
    /** @inheritdoc */
    send(recipient) {
        if (!recipient.tokenName) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid missing token name');
        }
        const tokenId = (0, utils_1.getHederaTokenIdFromName)(recipient.tokenName);
        if (!tokenId) {
            throw new sdk_core_1.InvalidParameterValueError(`Invalid token name: ${recipient.tokenName}`);
        }
        if (this._tokenName && this._tokenName !== recipient.tokenName) {
            throw new sdk_core_1.InvalidParameterValueError(`Invalid token: received ${recipient.tokenName} for ${this._tokenName} tx`);
        }
        this._tokenName = recipient.tokenName;
        return super.send(recipient);
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
