// @bitgo/sdk-coin-hbar/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toUint8Array = exports.toHex = void 0;
exports.isValidAddress = isValidAddress;
exports.isValidTransactionId = isValidTransactionId;
exports.isValidPublicKey = isValidPublicKey;
exports.isNodeEnvironment = isNodeEnvironment;
exports.getCurrentTime = getCurrentTime;
exports.isValidTimeString = isValidTimeString;
exports.isValidAmount = isValidAmount;
exports.isValidRawTransactionFormat = isValidRawTransactionFormat;
exports.stringifyAccountId = stringifyAccountId;
exports.stringifyTokenId = stringifyTokenId;
exports.stringifyTxTime = stringifyTxTime;
exports.removePrefix = removePrefix;
exports.isValidMemo = isValidMemo;
exports.createRawKey = createRawKey;
exports.convertFromStellarPub = convertFromStellarPub;
exports.isSameBaseAddress = isSameBaseAddress;
exports.getBaseAddress = getBaseAddress;
exports.getAddressDetails = getAddressDetails;
exports.normalizeAddress = normalizeAddress;
exports.isValidAddressWithPaymentId = isValidAddressWithPaymentId;
exports.buildHederaTokenID = buildHederaTokenID;
exports.buildHederaAccountID = buildHederaAccountID;
exports.isValidHederaTokenID = isValidHederaTokenID;
exports.getHederaTokenIdFromName = getHederaTokenIdFromName;
exports.getHederaTokenNameFromId = getHederaTokenNameFromId;
exports.isTokenTransfer = isTokenTransfer;
exports.validateStartTime = validateStartTime;
exports.normalizeStarttime = normalizeStarttime;
exports.sleep = sleep;
exports.shouldBroadcastNow = shouldBroadcastNow;
const _ = __importStar(require("lodash"));
const sdk_1 = require("@hashgraph/sdk");
const proto_1 = require("@hashgraph/proto");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const stellar = __importStar(require("stellar-sdk"));
const url_1 = __importDefault(require("url"));
const sdk_core_1 = require("@bitgo/sdk-core");
Object.defineProperty(exports, "toHex", { enumerable: true, get: function () { return sdk_core_1.toHex; } });
Object.defineProperty(exports, "toUint8Array", { enumerable: true, get: function () { return sdk_core_1.toUint8Array; } });
const statics_1 = require("@bitgo/statics");
const MAX_TINYBARS_AMOUNT = new bignumber_js_1.default(2).pow(63).minus(1);
/**
 * Returns whether the string is a valid Hedera account address
 *
 * In any form, `shard` and `realm` are assumed to be 0 if not provided.
 *
 * @param {string} address - The address to be validated
 * @returns {boolean} - The validation result
 */
function isValidAddress(address) {
    const addressArray = address.split('?memoId=');
    if (_.isEmpty(address) ||
        ![1, 2].includes(addressArray.length) ||
        !addressArray[0].match(/^\d+(?:(?=\.)(\.\d+){2}|(?!\.))$/) ||
        (addressArray[1] && !isValidMemo(addressArray[1]))) {
        return false;
    }
    try {
        const acc = sdk_1.AccountId.fromString(addressArray[0]);
        return !_.isNaN(acc.num);
    }
    catch (e) {
        return false;
    }
}
/**
 * Returns whether the string is a valid Hedera transaction id
 *
 * @param {string} txId - The transaction id to be validated
 * @returns {boolean} - The validation result
 */
function isValidTransactionId(txId) {
    if (_.isEmpty(txId)) {
        return false;
    }
    try {
        const tx = sdk_1.TransactionId.fromString(txId);
        if (_.isNil(tx.accountId)) {
            return false;
        }
        return !_.isNaN(tx.accountId.num);
    }
    catch (e) {
        return false;
    }
}
/**
 Returns whether the string is a valid Hedera public key
 *
 * @param {string} key - The public key to be validated
 * @returns {boolean} - The validation result
 */
function isValidPublicKey(key) {
    if (_.isEmpty(key)) {
        return false;
    }
    try {
        const pubKey = sdk_1.PublicKey.fromString(key.toLowerCase());
        return !_.isNaN(pubKey.toString());
    }
    catch (e) {
        return false;
    }
}
/**
 * Checks whether nodeJS.process exist and if a node version is defined to determine if this is an nodeJS environment
 *
 * @returns {boolean} - The validation result
 */
function isNodeEnvironment() {
    return typeof process !== 'undefined' && typeof process.versions.node !== 'undefined';
}
/**
 * Calculate the current time with nanoseconds precision
 *
 * @returns {string} - The current time in seconds
 */
function getCurrentTime() {
    if (isNodeEnvironment()) {
        const nanos = process.hrtime()[1];
        const seconds = (Date.now() * 1000000 + nanos) / 1000000000;
        return seconds.toFixed(9);
    }
    else {
        return (performance.timeOrigin + performance.now()).toFixed(9);
    }
}
/**
 * Returns whether the string is a valid timestamp
 *
 * Nanoseconds are optional and can be passed after a dot, for example: 1595374723.356981689
 *
 * @param {string} time - The timestamp to be validated
 * @returns {boolean} - The validation result
 */
function isValidTimeString(time) {
    return /^\d+(\.\d+)?$/.test(time);
}
/**
 * Returns whether the string is a valid amount number
 *
 * @param {string} amount - The string to validate
 * @returns {boolean} - The validation result
 */
function isValidAmount(amount) {
    const bigNumberAmount = new bignumber_js_1.default(amount);
    return (bigNumberAmount.isInteger() &&
        bigNumberAmount.isGreaterThanOrEqualTo(0) &&
        bigNumberAmount.isLessThanOrEqualTo(MAX_TINYBARS_AMOUNT));
}
/**
 * Returns whether the provided raw transaction accommodates to bitgo's preferred format
 *
 * @param {any} rawTransaction - The raw transaction to be checked
 * @returns {boolean} - The validation result
 */
function isValidRawTransactionFormat(rawTransaction) {
    const isAlphaNumeric = typeof rawTransaction === 'string' && /^[\da-fA-F]+$/.test(rawTransaction);
    const isValidBuffer = Buffer.isBuffer(rawTransaction) && !!Uint8Array.from(rawTransaction);
    return isAlphaNumeric || isValidBuffer;
}
/**
 * Returns a string representation of an {proto.IAccountID} object
 *
 * @param {proto.IAccountID} accountId - Account id to be cast to string
 * @returns {string} - The string representation of the {proto.IAccountID}
 */
function stringifyAccountId({ shardNum, realmNum, accountNum }) {
    return `${shardNum || 0}.${realmNum || 0}.${accountNum}`;
}
/**
 * Returns a string representation of an {proto.ITokenID} object
 *
 * @param {proto.ITokenID} - token id to be cast to string
 * @returns {string} - the string representation of the {proto.ITokenID}
 */
function stringifyTokenId({ shardNum, realmNum, tokenNum }) {
    return `${shardNum || 0}.${realmNum || 0}.${tokenNum}`;
}
/**
 * Returns a string representation of an {proto.ITimestamp} object
 *
 * @param {proto.ITimestamp} timestamp - Timestamp to be cast to string
 * @returns {string} - The string representation of the {proto.ITimestamp}
 */
function stringifyTxTime({ seconds, nanos }) {
    return `${seconds}.${nanos}`;
}
/**
 * Remove the specified prefix from a string only if it starts with that prefix
 *
 * @param {string} prefix - The prefix to be removed
 * @param {string} key - The original string, usually a private or public key
 * @returns {string} - The string without prefix
 */
function removePrefix(prefix, key) {
    if (key.startsWith(prefix)) {
        return key.slice(prefix.length);
    }
    return key;
}
/**
 * Check if this is a valid memo
 *
 * @param {string} memo
 * @returns {boolean}
 */
function isValidMemo(memo) {
    return !(_.isEmpty(memo) || Buffer.from(memo).length > 100);
}
/**
 * Uses the native hashgraph SDK function to get a raw key.
 *
 * @param {string} prv - Private key
 * @returns {PrivateKey}
 */
function createRawKey(prv) {
    return sdk_1.PrivateKey.fromString(prv);
}
/**
 * Converts a stellar public key to ed25519 hex format
 *
 * @param {string} stellarPub
 * @returns {string}
 */
function convertFromStellarPub(stellarPub) {
    if (!stellar.StrKey.isValidEd25519PublicKey(stellarPub)) {
        throw new Error('Not a valid stellar pub.');
    }
    const rawKey = stellar.StrKey.decodeEd25519PublicKey(stellarPub);
    return rawKey.toString('hex');
}
/**
 * Checks if two addresses have the same base address
 *
 * @param {String} address
 * @param {String} baseAddress
 * @returns {boolean}
 */
function isSameBaseAddress(address, baseAddress) {
    if (!isValidAddressWithPaymentId(address)) {
        throw new sdk_core_1.UtilsError(`invalid address: ${address}`);
    }
    return getBaseAddress(address) === getBaseAddress(baseAddress);
}
/**
 * Returns the base address portion of an address
 *
 * @param {String} address
 * @returns {String} - the base address
 */
function getBaseAddress(address) {
    const addressDetails = getAddressDetails(address);
    return addressDetails.address;
}
/**
 * Process address into address and memo id
 *
 * @param {string} rawAddress
 * @returns {AddressDetails} - object containing address and memo id
 */
function getAddressDetails(rawAddress) {
    const addressDetails = url_1.default.parse(rawAddress);
    const queryDetails = addressDetails.query ? new URLSearchParams(addressDetails.query) : undefined;
    const baseAddress = addressDetails.pathname;
    if (!isValidAddress(baseAddress)) {
        throw new sdk_core_1.UtilsError(`invalid address: ${rawAddress}`);
    }
    // address doesn't have a memo id or memoId is empty
    if (baseAddress === rawAddress) {
        return {
            address: rawAddress,
            memoId: undefined,
        };
    }
    if (!queryDetails || _.isNil(queryDetails.get('memoId'))) {
        // if there are more properties, the query details need to contain the memo id property
        throw new sdk_core_1.UtilsError(`invalid address with memo id: ${rawAddress}`);
    }
    const memoId = queryDetails.get('memoId');
    if (!isValidMemo(memoId)) {
        throw new sdk_core_1.UtilsError(`invalid address: '${rawAddress}', memoId is not valid`);
    }
    return {
        address: baseAddress,
        memoId,
    };
}
/**
 * Validate and return address with appended memo id
 *
 * @param {AddressDetails} addressDetails - Address which to append memo id
 * @returns {string} - Address with appended memo id
 */
function normalizeAddress({ address, memoId }) {
    if (memoId && isValidMemo(memoId)) {
        return `${address}?memoId=${memoId}`;
    }
    return address;
}
/**
 * Return boolean indicating whether input is a valid address with memo id
 *
 * @param {string} address - Address in the form <address>?memoId=<memoId>
 * @returns {boolean} - True if input is a valid address
 */
function isValidAddressWithPaymentId(address) {
    try {
        const addressDetails = getAddressDetails(address);
        return address === normalizeAddress(addressDetails);
    }
    catch (e) {
        return false;
    }
}
/**
 * Build hedera {proto.TokenID} object from token ID string
 *
 * @param {string} tokenID - The token ID to build
 * @returns {proto.TokenID} - The resulting proto TokenID object
 */
function buildHederaTokenID(tokenID) {
    const tokenData = sdk_1.TokenId.fromString(tokenID);
    return new proto_1.proto.TokenID({
        tokenNum: tokenData.num,
        realmNum: tokenData.realm,
        shardNum: tokenData.shard,
    });
}
/**
 * Build hedera {proto.AccountID} object from account ID string
 *
 * @param {string} accountID - The account ID to build
 * @returns {proto} - The resulting proto AccountID object
 */
function buildHederaAccountID(accountID) {
    const accountId = sdk_1.AccountId.fromString(accountID);
    return new proto_1.proto.AccountID({
        shardNum: accountId.shard,
        realmNum: accountId.realm,
        accountNum: accountId.num,
    });
}
/**
 * Check if Hedera token ID is valid and supported
 *
 * @param {string} tokenId - The token ID to validate
 * @returns {boolean} - True if tokenId is valid and supported
 */
function isValidHederaTokenID(tokenId) {
    const isFormatValid = !_.isEmpty(tokenId) && !!tokenId.match(/^\d+(?:(?=\.)(\.\d+){2}|(?!\.))$/);
    const isTokenSupported = getHederaTokenNameFromId(tokenId) !== undefined;
    return isFormatValid && isTokenSupported;
}
/**
 * Get the associated hedera token ID from token name, if supported
 *
 * @param {string} tokenName - The hedera token name
 * @returns {boolean} - The associated token ID or undefined if not supported
 */
function getHederaTokenIdFromName(tokenName) {
    if (statics_1.coins.has(tokenName)) {
        const token = statics_1.coins.get(tokenName);
        if (token.isToken && token instanceof statics_1.HederaToken) {
            return token.tokenId;
        }
    }
    return undefined;
}
/**
 * Get the associated hedera token from token ID, if supported
 *
 * @param tokenId - The token address
 * @returns {BaseCoin} - BaseCoin object for the matching token
 */
function getHederaTokenNameFromId(tokenId) {
    const tokensArray = statics_1.coins
        .filter((coin) => {
        return coin instanceof statics_1.HederaToken && coin.tokenId === tokenId;
    })
        .map((token) => token); // flatten coin map to array
    return tokensArray.length > 0 ? tokensArray[0] : undefined;
}
/**
 * Return boolean indicating whether input is a valid token transfer transaction
 *
 * @param {proto.ICryptoTransferTransactionBody | null} transferTxBody is a transfer transaction body
 * @returns {boolean} true is input is a valid token transfer transaction
 */
function isTokenTransfer(transferTxBody) {
    return !!transferTxBody && !!transferTxBody.tokenTransfers && transferTxBody.tokenTransfers.length > 0;
}
/** validates a startTime string to be a valid timestamp and in the future
 * @param {string} startTime - The startTime to be validated
 * @throws {Error} - if startTime is not a valid timestamp or is in the past
 * @returns {void}
 * */
function validateStartTime(startTime) {
    if (!isValidTimeString(startTime)) {
        throw new Error('invalid startTime, got: ' + startTime);
    }
    const currentTime = getCurrentTime();
    const startTimeFixed = normalizeStarttime(startTime);
    const result = new bignumber_js_1.default(startTimeFixed).isLessThanOrEqualTo(currentTime);
    if (result) {
        throw new Error('startTime must be a future timestamp, got: ' + startTime);
    }
}
function normalizeStarttime(startTime) {
    return new bignumber_js_1.default(startTime).toFixed(9);
}
/**
 * Await for a given amount of time in milliseconds
 * @param ms - The amount of time to wait in milliseconds
 * @returns {Promise<void>} - A promise that resolves after the given amount of time
 */
function sleep(ms) {
    console.log(`sleeping for ${ms} ms`);
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Check if the startTime is within the broadcast window (5 seconds after and 175 seconds after the startTime)
 */
function shouldBroadcastNow(startTime) {
    const startTimeFixed = normalizeStarttime(startTime);
    const currentTime = getCurrentTime();
    // startTime plus 5 seconds
    const startingTimeWindow = new bignumber_js_1.default(startTimeFixed).plus(5).toFixed(9);
    // startTime plus 170 seconds
    const endingTimeWindow = new bignumber_js_1.default(startTimeFixed).plus(175).toFixed(9);
    if (new bignumber_js_1.default(currentTime).isGreaterThan(endingTimeWindow)) {
        throw new Error('startTime window expired, got: ' +
            startTimeFixed +
            ' - currentTime: ' +
            currentTime +
            ' - endingTimeWindow ' +
            endingTimeWindow);
    }
    return (new bignumber_js_1.default(currentTime).isGreaterThanOrEqualTo(startingTimeWindow) &&
        new bignumber_js_1.default(currentTime).isLessThanOrEqualTo(endingTimeWindow));
}
