// @bitgo/sdk-coin-eos/dist/src/lib/transactionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    fromImplementation(rawTransaction) {
        throw new Error('Method not implemented.');
    }
    signImplementation(key) {
        throw new Error('Method not implemented.');
    }
    buildImplementation() {
        throw new Error('Method not implemented.');
    }
    validateKey(key) {
        throw new Error('Method not implemented.');
    }
    validateAddress(address, addressFormat) {
        throw new Error('Method not implemented.');
    }
    validateValue(value) {
        throw new Error('Method not implemented.');
    }
    validateRawTransaction(rawTransaction) {
        throw new Error('Method not implemented.');
    }
    validateTransaction(transaction) {
        throw new Error('Method not implemented.');
    }
    get transaction() {
        throw new Error('Method not implemented.');
    }
    set transaction(transaction) {
        throw new Error('Method not implemented.');
    }
}
exports.TransactionBuilder = TransactionBuilder;
