// @bitgo/abstract-utxo/dist/esm/transaction/descriptor/explainPsbt.js
import { descriptorWallet } from '@bitgo/wasm-utxo';
import { sumValues } from '../../wasmUtil';
function toRecipient(output, coinName) {
    const address = output.address ?? `scriptPubKey:${Buffer.from(output.script).toString('hex')}`;
    return {
        address,
        amount: output.value.toString(),
    };
}
function getInputSignaturesForInputIndex(psbt, inputIndex) {
    if (!psbt.hasPartialSignatures(inputIndex)) {
        return 0;
    }
    const partialSigs = psbt.getPartialSignatures(inputIndex);
    return partialSigs.reduce((agg, p) => {
        const valid = psbt.validateSignatureAtInput(inputIndex, p.pubkey);
        return agg + (valid ? 1 : 0);
    }, 0);
}
function getInputSignatures(psbt) {
    return Array.from({ length: psbt.inputCount() }, (_, i) => getInputSignaturesForInputIndex(psbt, i));
}
export function explainPsbt(psbt, descriptors, coinName) {
    const parsedTransaction = descriptorWallet.parse(psbt, descriptors, coinName);
    const { inputs, outputs } = parsedTransaction;
    const externalOutputs = outputs.filter((o) => o.scriptId === undefined);
    const changeOutputs = outputs.filter((o) => o.scriptId !== undefined);
    const fee = sumValues(inputs) - sumValues(outputs);
    const inputSignatures = getInputSignatures(psbt);
    return {
        inputSignatures,
        signatures: inputSignatures.reduce((a, b) => Math.min(a, b), Infinity),
        locktime: psbt.lockTime(),
        id: psbt.unsignedTxId(),
        outputs: externalOutputs.map((o) => toRecipient(o, coinName)),
        outputAmount: sumValues(externalOutputs).toString(),
        changeOutputs: changeOutputs.map((o) => toRecipient(o, coinName)),
        changeAmount: sumValues(changeOutputs).toString(),
        fee: fee.toString(),
    };
}
