// @bitgo/abstract-utxo/dist/esm/transaction/descriptor/parse.js
import { descriptorWallet } from '@bitgo/wasm-utxo';
import { getKeySignatures, toBip32Triple, UtxoNamedKeychains } from '../../keychains';
import { getDescriptorMapFromWallet, getPolicyForEnv } from '../../descriptor';
import { fromExtendedAddressFormatToScript, toExtendedAddressFormat } from '../recipient';
import { outputDifferencesWithExpected } from '../outputDifference';
import { sumValues, toWasmPsbt } from '../../wasmUtil';
function toRecipientOutput(recipient, coinName) {
    return {
        address: recipient.address,
        value: recipient.amount === 'max' ? 'max' : BigInt(recipient.amount),
        script: fromExtendedAddressFormatToScript(recipient.address, coinName),
        scriptId: undefined, // Recipients are external outputs
    };
}
function parseOutputsWithPsbt(psbt, descriptorMap, recipientOutputs, coinName) {
    const parsed = descriptorWallet.parse(psbt, descriptorMap, coinName);
    const outputs = parsed.outputs.map((output) => ({
        ...output,
        script: Buffer.from(output.script),
    }));
    const changeOutputs = outputs.filter((o) => o.scriptId !== undefined);
    const outputDiffs = outputDifferencesWithExpected(outputs, recipientOutputs);
    return {
        outputs,
        changeOutputs,
        ...outputDiffs,
    };
}
function toBaseOutputs(outputs, coinName) {
    return outputs.map((o) => ({
        address: toExtendedAddressFormat(o.script, coinName),
        amount: o.value === 'max' ? 'max' : BigInt(o.value),
        external: o.scriptId === undefined,
    }));
}
function toBaseParsedTransactionOutputs({ outputs, changeOutputs, explicitOutputs, implicitOutputs, missingOutputs }, coinName) {
    const explicitExternalOutputs = explicitOutputs.filter((o) => o.scriptId === undefined);
    const implicitExternalOutputs = implicitOutputs.filter((o) => o.scriptId === undefined);
    return {
        outputs: toBaseOutputs(outputs, coinName),
        changeOutputs: toBaseOutputs(changeOutputs, coinName),
        explicitExternalOutputs: toBaseOutputs(explicitExternalOutputs, coinName),
        explicitExternalSpendAmount: sumValues(explicitExternalOutputs),
        implicitExternalOutputs: toBaseOutputs(implicitExternalOutputs, coinName),
        implicitExternalSpendAmount: sumValues(implicitExternalOutputs),
        missingOutputs: toBaseOutputs(missingOutputs, coinName),
    };
}
export function toBaseParsedTransactionOutputsFromPsbt(psbt, descriptorMap, recipients, coinName) {
    const wasmPsbt = toWasmPsbt(psbt);
    return toBaseParsedTransactionOutputs(parseOutputsWithPsbt(wasmPsbt, descriptorMap, recipients.map((r) => toRecipientOutput(r, coinName)), coinName), coinName);
}
export function parse(coin, wallet, params) {
    if (params.txParams.allowExternalChangeAddress) {
        throw new Error('allowExternalChangeAddress is not supported for descriptor wallets');
    }
    if (params.txParams.changeAddress) {
        throw new Error('changeAddress is not supported for descriptor wallets');
    }
    const keychains = params.verification?.keychains;
    if (!keychains || !UtxoNamedKeychains.is(keychains)) {
        throw new Error('keychain is required for descriptor wallets');
    }
    const { recipients } = params.txParams;
    if (!recipients) {
        throw new Error('recipients is required');
    }
    const psbt = coin.decodeTransactionFromPrebuild(params.txPrebuild);
    let wasmPsbt;
    try {
        wasmPsbt = toWasmPsbt(psbt);
    }
    catch (e) {
        throw new Error(`expected psbt to be a wasm-utxo or utxo-lib PSBT: ${e instanceof Error ? e.message : e}`);
    }
    const walletKeys = toBip32Triple(keychains);
    const descriptorMap = getDescriptorMapFromWallet(wallet, walletKeys, getPolicyForEnv(params.wallet.bitgo.env));
    return {
        ...toBaseParsedTransactionOutputsFromPsbt(wasmPsbt, descriptorMap, recipients, coin.name),
        keychains,
        keySignatures: getKeySignatures(wallet) ?? {},
        customChange: undefined,
        needsCustomChangeKeySignatureVerification: false,
    };
}
