// @bitgo/abstract-utxo/dist/esm/transaction/descriptor/parseToAmountType.js
import { parse } from './parse';
export function toAmountType(v, t) {
    if (v === 'max') {
        return v;
    }
    switch (t) {
        case 'number':
            return Number(v);
        case 'bigint':
            return BigInt(v);
        case 'string':
            return String(v);
    }
}
function baseOutputToTNumber(output, amountType) {
    return {
        address: output.address,
        amount: toAmountType(output.amount, amountType),
        external: output.external,
    };
}
function entryToTNumber(k, v, params) {
    switch (k) {
        case 'outputs':
        case 'changeOutputs':
        case 'explicitExternalOutputs':
        case 'implicitExternalOutputs':
        case 'missingOutputs':
            if (v === undefined) {
                return [k, v];
            }
            if (Array.isArray(v)) {
                return [k, v.map((o) => baseOutputToTNumber(o, params.amountTypeBaseOutput))];
            }
            throw new Error('expected array');
        case 'explicitExternalSpendAmount':
        case 'implicitExternalSpendAmount':
            if (typeof v !== 'bigint') {
                throw new Error('expected bigint');
            }
            return [k, toAmountType(v, params.amountTypeAggregate)];
        default:
            return [k, v];
    }
}
export function parsedDescriptorTransactionToTNumber(obj, params) {
    return Object.fromEntries(Object.entries(obj).map(([k, v]) => entryToTNumber(k, v, params)));
}
export function parseToAmountType(coin, wallet, params) {
    return parsedDescriptorTransactionToTNumber(parse(coin, wallet, params), {
        amountTypeAggregate: coin.amountType,
        amountTypeBaseOutput: 'string',
    });
}
