// @bitgo/abstract-utxo/dist/esm/transaction/descriptor/verifyTransaction.js
import { TxIntentMismatchError } from '@bitgo/sdk-core';
import { toWasmPsbt } from '../../wasmUtil';
import { toBaseParsedTransactionOutputsFromPsbt } from './parse';
export class ValidationError extends Error {
    constructor(message) {
        super(message);
    }
}
export class ErrorMissingOutputs extends ValidationError {
    constructor(missingOutputs) {
        super(`missing outputs (count=${missingOutputs.length})`);
        this.missingOutputs = missingOutputs;
    }
}
export class ErrorImplicitExternalOutputs extends ValidationError {
    constructor(implicitExternalOutputs) {
        super(`unexpected implicit external outputs (count=${implicitExternalOutputs.length})`);
        this.implicitExternalOutputs = implicitExternalOutputs;
    }
}
export class AggregateValidationError extends ValidationError {
    constructor(errors) {
        super(`aggregate validation error (count=${errors.length})`);
        this.errors = errors;
    }
}
export function assertExpectedOutputDifference(parsedOutputs) {
    const errors = [];
    if (parsedOutputs.missingOutputs.length > 0) {
        errors.push(new ErrorMissingOutputs(parsedOutputs.missingOutputs));
    }
    if (parsedOutputs.implicitExternalOutputs.length > 0) {
        // FIXME: for paygo we need to relax this a little bit
        errors.push(new ErrorImplicitExternalOutputs(parsedOutputs.implicitExternalOutputs));
    }
    if (errors.length > 0) {
        // FIXME(BTC-1688): enable ES2021
        // throw new AggregateError(errors);
        throw new AggregateValidationError(errors);
    }
}
export function assertValidTransaction(psbt, descriptors, recipients, coinName) {
    assertExpectedOutputDifference(toBaseParsedTransactionOutputsFromPsbt(psbt, descriptors, recipients, coinName));
}
/**
 * Wrapper around assertValidTransaction that returns a boolean instead of throwing.
 *
 * We follow the AbstractUtxoCoin interface here which is a bit confused - the return value is a boolean but we
 * also throw errors (because we actually want to know what went wrong).
 *
 * @param coin
 * @param params
 * @param descriptorMap
 * @returns {boolean} True if verification passes
 * @throws {TxIntentMismatchError} if transaction validation fails
 */
export async function verifyTransaction(coin, params, descriptorMap) {
    const tx = coin.decodeTransactionFromPrebuild(params.txPrebuild);
    let psbt;
    try {
        psbt = toWasmPsbt(tx);
    }
    catch (e) {
        const txExplanation = await TxIntentMismatchError.tryGetTxExplanation(coin, params.txPrebuild);
        const errorDetail = e instanceof Error ? e.message : String(e);
        throw new TxIntentMismatchError(`unexpected transaction type: ${errorDetail}`, params.reqId, [params.txParams], params.txPrebuild.txHex, txExplanation);
    }
    assertValidTransaction(psbt, descriptorMap, params.txParams.recipients ?? [], coin.name);
    return true;
}
