// @babylonlabs-io/babylon-proto-ts/dist/generated/babylon/incentive/params.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.6.1
//   protoc               unknown
// source: babylon/incentive/params.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "babylon.incentive";
function createBaseParams() {
    return { btcStakingPortion: "" };
}
exports.Params = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.btcStakingPortion !== "") {
            writer.uint32(10).string(message.btcStakingPortion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.btcStakingPortion = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { btcStakingPortion: isSet(object.btcStakingPortion) ? globalThis.String(object.btcStakingPortion) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.btcStakingPortion !== "") {
            obj.btcStakingPortion = message.btcStakingPortion;
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParams();
        message.btcStakingPortion = (_a = object.btcStakingPortion) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
