// @bitgo/abstract-utxo/dist/esm/transaction/fixedScript/signTransaction.js
import assert from 'assert';
import { isTriple } from '@bitgo/sdk-core';
import _ from 'lodash';
import { bitgo } from '@bitgo/utxo-lib';
import { fixedScriptWallet } from '@bitgo/wasm-utxo';
import { toUtxolibBIP32 } from '../../wasmUtil';
import { signLegacyTransaction } from './signLegacyTransaction';
import { signPsbtWithMusig2ParticipantUtxolib, signAndVerifyPsbt as signAndVerifyPsbtUtxolib } from './signPsbtUtxolib';
import { signPsbtWithMusig2ParticipantWasm, signAndVerifyPsbtWasm } from './signPsbtWasm';
import { getReplayProtectionPubkeys } from './replayProtection';
export function signAndVerifyPsbt(psbt, signerKeychain, rootWalletKeys, replayProtection) {
    if (psbt instanceof bitgo.UtxoPsbt) {
        return signAndVerifyPsbtUtxolib(psbt, toUtxolibBIP32(signerKeychain));
    }
    assert(rootWalletKeys, 'rootWalletKeys required for wasm-utxo signing');
    assert(replayProtection, 'replayProtection required for wasm-utxo signing');
    return signAndVerifyPsbtWasm(psbt, signerKeychain, rootWalletKeys, replayProtection);
}
export async function signTransaction(coin, tx, signerKeychain, coinName, params) {
    let isLastSignature = false;
    if (_.isBoolean(params.isLastSignature)) {
        // if build is called instead of buildIncomplete, no signature placeholders are left in the sig script
        isLastSignature = params.isLastSignature;
    }
    if (tx instanceof bitgo.UtxoPsbt) {
        const signedPsbt = await signPsbtWithMusig2ParticipantUtxolib(coin, tx, signerKeychain ? toUtxolibBIP32(signerKeychain) : undefined, {
            signingStep: params.signingStep,
            walletId: params.walletId,
        });
        if (isLastSignature) {
            signedPsbt.finalizeAllInputs();
            return signedPsbt.extractTransaction();
        }
        return signedPsbt;
    }
    else if (tx instanceof fixedScriptWallet.BitGoPsbt) {
        assert(params.pubs, 'pubs are required for fixed script signing');
        assert(isTriple(params.pubs), 'pubs must be a triple');
        const rootWalletKeys = fixedScriptWallet.RootWalletKeys.fromXpubs(params.pubs);
        const signedPsbt = await signPsbtWithMusig2ParticipantWasm(coin, tx, signerKeychain, rootWalletKeys, {
            replayProtection: {
                publicKeys: getReplayProtectionPubkeys(coinName),
            },
            signingStep: params.signingStep,
            walletId: params.walletId,
        });
        if (isLastSignature) {
            signedPsbt.finalizeAllInputs();
            return Buffer.from(signedPsbt.extractTransaction().toBytes());
        }
        return signedPsbt;
    }
    return signLegacyTransaction(tx, signerKeychain, coinName, {
        isLastSignature,
        signingStep: params.signingStep,
        txInfo: params.txInfo,
        pubs: params.pubs,
        cosignerPub: params.cosignerPub,
    });
}
