// @bitgo/sdk-core/dist/src/bitgo/staking/goStakingWallet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoStakingWallet = void 0;
const assert_1 = __importDefault(require("assert"));
class GoStakingWallet {
    constructor(wallet) {
        this.wallet = wallet;
        this.bitgo = wallet.bitgo;
    }
    get accountId() {
        return this.wallet.id();
    }
    async stake(options) {
        // call preview
        const preview = await this.previewStake({
            amount: options.amount,
            clientId: options.clientId,
        });
        // sign the transaction
        const halfSignedTransaction = (await this.wallet.prebuildAndSignTransaction({
            walletPassphrase: options.walletPassphrase,
            prebuildTx: {
                payload: preview.payload,
            },
        }));
        // call finalize to submit the go staking request to go staking service
        (0, assert_1.default)(halfSignedTransaction.halfSigned?.payload, 'missing payload in half signed transaction');
        const finalOptions = {
            amount: options.amount,
            clientId: options.clientId,
            frontTransferSendRequest: {
                halfSigned: {
                    payload: halfSignedTransaction.halfSigned.payload,
                },
            },
        };
        return (await this.finalizeStake(finalOptions));
    }
    /**
     * Unstake request
     * @param options
     */
    async unstake(options) {
        return (await this.createGoStakingRequest(options, 'finalize', 'UNSTAKE'));
    }
    /**
     * Preview staking request
     * @param options
     */
    async previewStake(options) {
        return (await this.createGoStakingRequest(options, 'preview', 'STAKE'));
    }
    /**
     * Finalize staking request
     * will prepare the payload and sign the transaction
     * and submit it to the go-staking-service
     * @param options
     */
    async finalizeStake(options) {
        return (await this.createGoStakingRequest(options, 'finalize', 'STAKE'));
    }
    /**
     * Get go staking request
     * @param goStakingRequestId
     */
    async getGoStakingRequest(goStakingRequestId) {
        return await this.bitgo.get(this.bitgo.microservicesUrl(this.getGoStakingRequestURL(goStakingRequestId))).result();
    }
    /**
     * Get go staking requests by coin
     * @param options
     */
    async getGoStakingRequestsByWalletCoin(options) {
        return await this.bitgo
            .get(this.bitgo.microservicesUrl(this.goStakingRequestBaseURL()))
            .query({
            ...options,
        })
            .result();
    }
    /**
     * Get all go staking requests for the account
     * @param options
     */
    async getGoStakingRequests(options = {}) {
        return await this.bitgo
            .get(this.bitgo.microservicesUrl(this.getGoStakingRequestsUrl()))
            .query({
            ...options,
        })
            .result();
    }
    /**
     * Get go staking wallets for the account
     * @param options
     */
    async getGoStakingWallets(options) {
        return await this.bitgo
            .get(this.bitgo.microservicesUrl(this.getGoStakingWalletsUrl()))
            .query({
            ...options,
        })
            .result();
    }
    /**
     * Get go staking wallets for the account
     */
    async getGoStakingWallet() {
        return await this.bitgo.get(this.bitgo.microservicesUrl(this.getGoStakingWalletUrl())).result();
    }
    async createGoStakingRequest(options, path, type) {
        return await this.bitgo
            .post(this.bitgo.microservicesUrl(`${this.goStakingRequestBaseURL()}/${path}`))
            .send({
            ...options,
            type: type,
        })
            .result();
    }
    goStakingBaseURL() {
        return `/api/go-staking/v1`;
    }
    goStakingRequestBaseURL() {
        return `${this.goStakingBaseURL()}/${this.wallet.baseCoin.getChain()}/accounts/${this.accountId}/requests`;
    }
    getGoStakingRequestURL(stakingRequestId) {
        return `${this.goStakingRequestBaseURL()}/${stakingRequestId}`;
    }
    getGoStakingRequestsUrl() {
        return `${this.goStakingBaseURL()}/accounts/${this.accountId}/requests`;
    }
    getGoStakingWalletsUrl() {
        return `${this.goStakingBaseURL()}/accounts/${this.accountId}/coins`;
    }
    getGoStakingWalletUrl() {
        return `${this.goStakingBaseURL()}/${this.wallet.baseCoin.getChain()}/accounts/${this.accountId}/coins`;
    }
}
exports.GoStakingWallet = GoStakingWallet;
