// @bitgo/sdk-core/dist/src/bitgo/staking/stakingWallet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakingWallet = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const isEqual_1 = __importDefault(require("lodash/isEqual"));
const utils_1 = require("../utils");
const assert_1 = __importDefault(require("assert"));
const transactionUtils_1 = require("../utils/transactionUtils");
const debug_1 = __importDefault(require("debug"));
const account_lib_1 = require("../../account-lib");
class StakingWallet {
    constructor(wallet, isEthTss) {
        this.wallet = wallet;
        this.bitgo = wallet.bitgo;
        this.tssUtil = new utils_1.TssUtils(this.bitgo, this.wallet.baseCoin, this.wallet);
        this.isEthTss = isEthTss;
    }
    get walletId() {
        return this.wallet.id();
    }
    get coin() {
        return this.wallet.baseCoin.tokenConfig ? this.wallet.baseCoin.tokenConfig.type : this.wallet.coin();
    }
    /**
     * Stake coins
     * @param options - stake options
     * @return StakingRequest
     */
    async stake(options) {
        return await this.createStakingRequest(options, 'STAKE');
    }
    /**
     * Unstake coins
     * @param options - unstake options
     * @return StakingRequest
     */
    async unstake(options) {
        return await this.createStakingRequest(options, 'UNSTAKE');
    }
    /**
     * Submit a request to switch the validator used for a specific delegation
     * This will create a new delegation with the new validator address and mark the old delegation as inactive
     * @param options - switch validator options
     * @return StakingRequest
     */
    async switchValidator(options) {
        return await this.createStakingRequest(options, 'SWITCH_VALIDATOR');
    }
    /**
     * Submit a request to claim rewards for a specific delegation
     * @param options - claim rewards options
     * @return StakingRequest
     */
    async claimRewards(options) {
        return await this.createStakingRequest(options, 'CLAIM_REWARDS');
    }
    /**
     * Cancel staking request
     * @param stakingRequestId - id of the staking request to cancel
     * @return StakingRequest
     */
    async cancelStakingRequest(stakingRequestId) {
        return await this.bitgo.del(this.bitgo.microservicesUrl(this.stakingRequestUrl(stakingRequestId))).result();
    }
    /**
     * Fetch delegations for a specific wallet
     * @param options - unstake options
     * @return StakingRequest
     */
    async delegations(options) {
        return await this.getDelegations(options);
    }
    /**
     * Get a staking request by ID
     * @param stakingRequestId - id of the staking request to retrieve
     * @return StakingRequest
     */
    async getStakingRequest(stakingRequestId) {
        return await this.bitgo.get(this.bitgo.microservicesUrl(this.stakingRequestUrl(stakingRequestId))).result();
    }
    /**
     * Get transactions ready to sign
     * @param stakingRequestId
     * @return TransactionsReadyToSign
     */
    async getTransactionsReadyToSign(stakingRequestId) {
        const stakingRequest = await this.getStakingRequest(stakingRequestId);
        const readyToSign = stakingRequest.transactions.filter((transaction) => transaction.status === `READY`);
        const newTransactions = stakingRequest.transactions.filter((transaction) => transaction.status === `NEW`);
        return Promise.resolve({
            allSigningComplete: stakingRequest.transactions.length > 0 && newTransactions.length === 0 && readyToSign.length === 0,
            transactions: readyToSign,
        });
    }
    /**
     * Build the staking transaction
     * If TSS delete signature shares, else expand build params and then build
     * @param transaction - staking transaction to build
     */
    async build(transaction) {
        if ((this.wallet.multisigType() === 'tss' && this.wallet.baseCoin.getFamily() !== 'eth') || this.isEthTss) {
            if (!transaction.txRequestId) {
                throw new Error('txRequestId is required to sign and send');
            }
            // delete signature shares before signing for transaction request API
            await this.tssUtil.deleteSignatureShares(transaction.txRequestId);
            return {
                transaction: transaction,
                result: {
                    walletId: this.walletId,
                    txRequestId: transaction.txRequestId,
                },
            };
        }
        else {
            transaction = await this.expandBuildParams(transaction);
            if (!transaction.buildParams) {
                throw Error(`Staking transaction ${transaction.id} build params not expanded`);
            }
            const wallet = await this.extractWallet(transaction);
            return {
                transaction: transaction,
                result: await wallet.prebuildTransaction(transaction.buildParams),
            };
        }
    }
    async extractWallet(transaction) {
        return this.isBtcUndelegate(transaction)
            ? await this.getDescriptorWallet(transaction)
            : await this.getWalletForBuildingAndSigning();
    }
    isBtcUndelegate(transaction) {
        return (this.wallet.baseCoin.getFamily() === 'btc' && transaction.transactionType.toLowerCase() === 'undelegate_withdraw');
    }
    isStx() {
        return this.wallet.baseCoin.getFamily() === 'stx';
    }
    isTrxStaking(transaction) {
        return this.wallet.baseCoin.getFamily() === 'trx';
    }
    /**
     * Sign the staking transaction
     * @param signOptions
     * @param stakingPrebuildTransaction
     */
    async sign(signOptions, stakingPrebuildTransaction) {
        const reqId = new utils_1.RequestTracer();
        const isBtcUndelegate = this.wallet.baseCoin.getFamily() === 'btc' &&
            stakingPrebuildTransaction.transaction.transactionType.toLowerCase() === 'undelegate_withdraw';
        const wallet = isBtcUndelegate
            ? await this.getDescriptorWallet(stakingPrebuildTransaction.transaction)
            : await this.getWalletForBuildingAndSigning();
        const keychain = await wallet.baseCoin.keychains().getKeysForSigning({
            wallet: this.wallet,
            reqId: reqId,
        });
        return {
            transaction: stakingPrebuildTransaction.transaction,
            signed: await wallet.signTransaction({
                txPrebuild: stakingPrebuildTransaction.result,
                walletPassphrase: signOptions.walletPassphrase,
                keychain: keychain[0],
            }),
        };
    }
    /**
     * Send the signed staking transaction if required. Send call is not required if api version is full
     * and this method will return the staking transaction from the incoming object.
     * @param signedTransaction
     */
    async send(signedTransaction) {
        if (this.isSendCallRequired()) {
            return await this.bitgo
                .post(this.bitgo.microservicesUrl(this.stakingTransactionURL(signedTransaction.transaction)))
                .send(signedTransaction.signed)
                .result();
        }
        return signedTransaction.transaction;
    }
    /**
     * @Deprecated use buildAndSign
     * Build, sign and send the transaction.
     * @param signOptions
     * @param transaction
     */
    async buildSignAndSend(signOptions, transaction) {
        return await this.buildAndSign(signOptions, transaction).then((result) => {
            return this.send(result);
        });
    }
    /**
     * Create prebuilt staking transaction.
     *
     * for transactions with tx request id (TSS transactions), we need to delete signature shares before creating prebuild transaction
     * we only need to get transaction build params if they exist to pre build
     *
     * @param transaction
     */
    async prebuildSelfManagedStakingTransaction(transaction) {
        if (transaction.txRequestId) {
            await this.tssUtil.deleteSignatureShares(transaction.txRequestId);
        }
        const buildParams = (await this.expandBuildParams(transaction)).buildParams;
        const formattedParams = {
            ...buildParams,
            coin: this.coin,
            walletId: this.walletId,
            walletType: this.wallet.type(),
            preview: true,
        };
        return await (await this.getWalletForBuildingAndSigning()).prebuildTransaction(formattedParams);
    }
    /**
     * Build and sign the transaction.
     * @param signOptions
     * @param transaction
     */
    async buildAndSign(signOptions, transaction) {
        const builtTx = await this.build(transaction);
        // default to verifying a transaction unless explicitly skipped
        // skipping the verification for btc undelegate because it is just single sig
        // TODO: SC-3183 (add trx staking verification)
        // TODO: SC-3508 (add STX staking verification)
        const skipVerification = (signOptions.transactionVerificationOptions?.skipTransactionVerification ||
            this.isBtcUndelegate(transaction) ||
            this.isTrxStaking(transaction) ||
            this.isStx()) ??
            false;
        if (!isStakingTxRequestPrebuildResult(builtTx.result) && !skipVerification) {
            await this.validateBuiltStakingTransaction(builtTx.transaction, builtTx);
        }
        return await this.sign(signOptions, builtTx);
    }
    async expandBuildParams(stakingTransaction) {
        return await this.bitgo
            .get(this.bitgo.microservicesUrl(this.stakingTransactionURL(stakingTransaction)))
            .query({ expandBuildParams: true })
            .result();
    }
    async createStakingRequest(options, type) {
        return await this.bitgo
            .post(this.bitgo.microservicesUrl(this.stakingRequestsURL()))
            .send({
            ...options,
            type: type,
        })
            .result();
    }
    stakingRequestsURL() {
        return `/api/staking/v1/${this.coin}/wallets/${this.walletId}/requests`;
    }
    async getDelegations(options) {
        return await this.bitgo.get(this.bitgo.microservicesUrl(this.stakingDelegationsURL())).query(options).result();
    }
    stakingDelegationsURL() {
        return `/api/staking/v1/${this.coin}/wallets/${this.walletId}/delegations`;
    }
    stakingRequestUrl(stakingRequestId) {
        return `${this.stakingRequestsURL()}/${stakingRequestId}`;
    }
    stakingTransactionURL(stakingTransaction) {
        return `${this.stakingRequestUrl(stakingTransaction.stakingRequestId)}/transactions/${stakingTransaction.id}`;
    }
    async getWalletForBuildingAndSigning() {
        if (this.wallet.baseCoin.tokenConfig) {
            if (!this.tokenParentWallet) {
                this.tokenParentWallet = await this.bitgo
                    .coin(this.wallet.baseCoin.tokenConfig.coin)
                    .wallets()
                    .get({ id: this.wallet.id() });
            }
            return this.tokenParentWallet;
        }
        else {
            return Promise.resolve(this.wallet);
        }
    }
    /**
     * Send API call is only required for TSS TxRequest api version lite or multi-sig transactions.
     * For Full api version, sign transaction moves the transaction to delivered state.
     * @returns true if send API call to staking service is required else false
     */
    isSendCallRequired() {
        if (this.wallet.baseCoin.getFamily() === statics_1.CoinFamily.ETH) {
            return !this.isEthTss;
        }
        else if (this.wallet.multisigType() === 'tss') {
            return this.wallet.baseCoin.getMPCAlgorithm() !== 'ecdsa';
        }
        else {
            return true;
        }
    }
    async getDescriptorWallet(transaction) {
        (0, assert_1.default)(transaction.buildParams?.senderWalletId, 'senderWalletId is required for btc undelegate transaction');
        return await this.wallet.baseCoin.wallets().get({ id: transaction.buildParams.senderWalletId });
    }
    async validateBuiltStakingTransaction(transaction, prebuiltStakingTransaction) {
        const { buildParams } = transaction;
        const { result } = prebuiltStakingTransaction;
        const coin = this.wallet.baseCoin;
        (0, debug_1.default)(`Validating staking transaction ${transaction.stakingRequestId} with prebuilt transaction`);
        if (!('txHex' in result) || !result.txHex) {
            (0, debug_1.default)(`Skipping validation for staking transaction ${transaction.stakingRequestId} - txHex is undefined`);
            return;
        }
        const explainedTransaction = await coin.explainTransaction(result, this.wallet);
        const mismatchErrors = [];
        if (buildParams?.recipients && buildParams.recipients.length > 0) {
            const userRecipientMap = new Map(buildParams.recipients.map((recipient) => [recipient.address.toLowerCase(), recipient]));
            const platformRecipientMap = new Map((explainedTransaction?.outputs ?? []).map((recipient) => [recipient.address.toLowerCase(), recipient]));
            for (const [address] of platformRecipientMap) {
                if (!userRecipientMap.has(address)) {
                    mismatchErrors.push(`Unexpected recipient address found in built transaction: ${address}`);
                }
            }
            for (const [address, userRecipient] of userRecipientMap) {
                const platformRecipient = platformRecipientMap.get(address);
                if (!platformRecipient) {
                    mismatchErrors.push(`Expected recipient address not found in built transaction: ${address}`);
                    continue;
                }
                const matchResult = (0, transactionUtils_1.transactionRecipientsMatch)(userRecipient, platformRecipient);
                if (!matchResult.amountMatch) {
                    mismatchErrors.push(`Recipient ${address} amount mismatch. Expected: ${userRecipient.amount}, Got: ${platformRecipient.amount}`);
                }
                if (!matchResult.tokenMatch) {
                    mismatchErrors.push(`Recipient ${address} token mismatch. Expected: ${userRecipient.tokenName ?? 'native coin'}, Got: ${platformRecipient.tokenName ?? 'native coin'}`);
                }
            }
        }
        if (buildParams?.memo && explainedTransaction.memo !== buildParams.memo) {
            mismatchErrors.push(`Memo mismatch. Expected: '${JSON.stringify(buildParams.memo)}', Got: '${JSON.stringify(explainedTransaction.memo)}'`);
        }
        if (buildParams?.gasLimit && String(explainedTransaction.gasLimit) !== String(buildParams.gasLimit)) {
            mismatchErrors.push(`Gas Limit mismatch. Expected: ${buildParams.gasLimit}, Got: ${explainedTransaction.gasLimit}`);
        }
        if (buildParams?.type &&
            explainedTransaction.type !== undefined &&
            account_lib_1.TransactionType[buildParams.type] !== explainedTransaction.type) {
            mismatchErrors.push(`Transaction type mismatch. Expected: '${buildParams.type}', Got: '${explainedTransaction.type}'`);
        }
        if (buildParams?.solInstructions) {
            if (!(0, isEqual_1.default)(explainedTransaction.solInstructions, buildParams.solInstructions)) {
                mismatchErrors.push(`Solana instructions mismatch. Expected: ${JSON.stringify(buildParams.solInstructions)}, Got: ${JSON.stringify(explainedTransaction.solInstructions)}`);
            }
        }
        if (mismatchErrors.length > 0) {
            const errorMessage = `Staking transaction validation failed before signing: ${mismatchErrors.join('; ')}`;
            (0, debug_1.default)(errorMessage);
            throw new Error(errorMessage);
        }
        if (!buildParams) {
            (0, debug_1.default)(`Cannot perform deep validation for staking transaction ${transaction.stakingRequestId} without specified build params`);
        }
    }
}
exports.StakingWallet = StakingWallet;
function isStakingTxRequestPrebuildResult(tx) {
    return tx.txRequestId !== undefined;
}
