// @bitgo/wasm-solana/dist/cjs/js/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.findWithdrawAuthorityProgramAddress = exports.getAssociatedTokenAddress = exports.sysvarRecentBlockhashes = exports.nonceAccountSpace = exports.stakeAccountSpace = exports.stakePoolProgramId = exports.ataProgramId = exports.token2022ProgramId = exports.tokenProgramId = exports.memoProgramId = exports.computeBudgetProgramId = exports.stakeProgramId = exports.systemProgramId = exports.buildTransactionFromIntent = exports.buildFromIntent = exports.buildFromVersionedData = exports.parseTransaction = exports.isVersionedTransaction = exports.VersionedTransaction = exports.Transaction = exports.Pubkey = exports.Keypair = exports.builder = exports.parser = exports.transaction = exports.pubkey = exports.keypair = void 0;
const wasm = __importStar(require("./wasm/wasm_solana.js"));
// Force webpack to include the WASM module
void wasm;
// Namespace exports for explicit imports
exports.keypair = __importStar(require("./keypair.js"));
exports.pubkey = __importStar(require("./pubkey.js"));
exports.transaction = __importStar(require("./transaction.js"));
exports.parser = __importStar(require("./parser.js"));
exports.builder = __importStar(require("./builder.js"));
// Top-level class exports for convenience
var keypair_js_1 = require("./keypair.js");
Object.defineProperty(exports, "Keypair", { enumerable: true, get: function () { return keypair_js_1.Keypair; } });
var pubkey_js_1 = require("./pubkey.js");
Object.defineProperty(exports, "Pubkey", { enumerable: true, get: function () { return pubkey_js_1.Pubkey; } });
var transaction_js_1 = require("./transaction.js");
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return transaction_js_1.Transaction; } });
// Versioned transaction support
var versioned_js_1 = require("./versioned.js");
Object.defineProperty(exports, "VersionedTransaction", { enumerable: true, get: function () { return versioned_js_1.VersionedTransaction; } });
Object.defineProperty(exports, "isVersionedTransaction", { enumerable: true, get: function () { return versioned_js_1.isVersionedTransaction; } });
// Top-level function exports
var parser_js_1 = require("./parser.js");
Object.defineProperty(exports, "parseTransaction", { enumerable: true, get: function () { return parser_js_1.parseTransaction; } });
var builder_js_1 = require("./builder.js");
Object.defineProperty(exports, "buildFromVersionedData", { enumerable: true, get: function () { return builder_js_1.buildFromVersionedData; } });
var intentBuilder_js_1 = require("./intentBuilder.js");
Object.defineProperty(exports, "buildFromIntent", { enumerable: true, get: function () { return intentBuilder_js_1.buildFromIntent; } });
Object.defineProperty(exports, "buildTransactionFromIntent", { enumerable: true, get: function () { return intentBuilder_js_1.buildFromIntent; } });
// Program ID constants (from WASM)
var wasm_solana_js_1 = require("./wasm/wasm_solana.js");
Object.defineProperty(exports, "systemProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.system_program_id; } });
Object.defineProperty(exports, "stakeProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.stake_program_id; } });
Object.defineProperty(exports, "computeBudgetProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.compute_budget_program_id; } });
Object.defineProperty(exports, "memoProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.memo_program_id; } });
Object.defineProperty(exports, "tokenProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.token_program_id; } });
Object.defineProperty(exports, "token2022ProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.token_2022_program_id; } });
Object.defineProperty(exports, "ataProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.ata_program_id; } });
Object.defineProperty(exports, "stakePoolProgramId", { enumerable: true, get: function () { return wasm_solana_js_1.stake_pool_program_id; } });
Object.defineProperty(exports, "stakeAccountSpace", { enumerable: true, get: function () { return wasm_solana_js_1.stake_account_space; } });
Object.defineProperty(exports, "nonceAccountSpace", { enumerable: true, get: function () { return wasm_solana_js_1.nonce_account_space; } });
// Sysvar addresses
Object.defineProperty(exports, "sysvarRecentBlockhashes", { enumerable: true, get: function () { return wasm_solana_js_1.sysvar_recent_blockhashes; } });
// PDA derivation functions (eliminates @solana/web3.js dependency)
Object.defineProperty(exports, "getAssociatedTokenAddress", { enumerable: true, get: function () { return wasm_solana_js_1.get_associated_token_address; } });
Object.defineProperty(exports, "findWithdrawAuthorityProgramAddress", { enumerable: true, get: function () { return wasm_solana_js_1.find_withdraw_authority_program_address; } });
