// @bitgo/wasm-solana/dist/cjs/js/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const wasm_solana_js_1 = require("./wasm/wasm_solana.js");
const pubkey_js_1 = require("./pubkey.js");
/**
 * Solana Transaction — deserialization wrapper for signing and serialization.
 *
 * Use `Transaction.fromBytes(bytes)` to create an instance for signing.
 * Use `parseTransaction(tx)` from parser.ts to get decoded instruction data.
 *
 * @example
 * ```typescript
 * import { Transaction, parseTransaction } from '@bitgo/wasm-solana';
 *
 * const tx = Transaction.fromBytes(txBytes);
 *
 * // Parse for decoded instructions
 * const parsed = parseTransaction(tx);
 * for (const instr of parsed.instructionsData) {
 *   if (instr.type === 'Transfer') {
 *     console.log(`${instr.amount} lamports to ${instr.toAddress}`);
 *   }
 * }
 *
 * // Sign and serialize
 * tx.addSignature(pubkey, signature);
 * const signedBytes = tx.toBytes();
 * ```
 */
class Transaction {
    _wasm;
    constructor(_wasm) {
        this._wasm = _wasm;
    }
    /**
     * Deserialize a transaction from raw bytes
     * @param bytes - The raw transaction bytes
     * @returns A Transaction instance
     */
    static fromBytes(bytes) {
        const wasm = wasm_solana_js_1.WasmTransaction.from_bytes(bytes);
        return new Transaction(wasm);
    }
    /**
     * Create a Transaction from a WasmTransaction instance.
     * @internal Used by builder functions
     */
    static fromWasm(wasm) {
        return new Transaction(wasm);
    }
    /**
     * Get the fee payer address as a base58 string
     * Returns null if there are no account keys (shouldn't happen for valid transactions)
     */
    get feePayer() {
        return this._wasm.fee_payer ?? null;
    }
    /**
     * Get the recent blockhash as a base58 string
     */
    get recentBlockhash() {
        return this._wasm.recent_blockhash;
    }
    /**
     * Get the number of signatures in the transaction
     */
    get numSignatures() {
        return this._wasm.num_signatures;
    }
    /**
     * Get the transaction ID (first signature as base58).
     *
     * For Solana, the transaction ID is the first signature.
     * Returns `undefined` if the transaction is unsigned (no signatures or all-zeros signature).
     *
     * @example
     * ```typescript
     * const tx = Transaction.fromBytes(txBytes);
     * tx.addSignature(pubkey, signature);
     * console.log(tx.id); // Base58 encoded signature
     * ```
     */
    get id() {
        return this._wasm.id;
    }
    /**
     * Get the signable message payload (what gets signed)
     * This is the serialized message that signers sign
     * @returns The message bytes
     */
    signablePayload() {
        return this._wasm.signable_payload();
    }
    /**
     * Serialize the message portion of the transaction.
     * Alias for signablePayload() - provides compatibility with @solana/web3.js API.
     * @returns The serialized message bytes
     */
    serializeMessage() {
        return this.signablePayload();
    }
    /**
     * Serialize the transaction to bytes
     * @returns The serialized transaction bytes
     */
    toBytes() {
        return this._wasm.to_bytes();
    }
    /**
     * Serialize to network broadcast format.
     * @returns The transaction as bytes ready for broadcast
     */
    toBroadcastFormat() {
        return this.toBytes();
    }
    /**
     * Get all account keys as Pubkey instances
     * @returns Array of account public keys
     */
    accountKeys() {
        const keys = Array.from(this._wasm.account_keys());
        return keys.map((k) => pubkey_js_1.Pubkey.fromBase58(k));
    }
    /**
     * Get all signatures as byte arrays.
     * Provides compatibility with @solana/web3.js Transaction.signatures API.
     * @returns Array of signature byte arrays
     */
    get signatures() {
        return Array.from(this._wasm.signatures());
    }
    /**
     * Get all signatures as byte arrays (method form).
     * Alias for the `signatures` property getter.
     * @returns Array of signature byte arrays
     */
    getSignatures() {
        return this.signatures;
    }
    /**
     * Get all instructions in the transaction.
     * Returns an array with programId, accounts, and data for each instruction.
     *
     * Note: This is a getter property to provide compatibility with code
     * expecting @solana/web3.js Transaction.instructions API. If you need
     * to call this as a method, use `getInstructions()` instead.
     */
    get instructions() {
        const rawInstructions = this._wasm.instructions();
        return Array.from(rawInstructions);
    }
    /**
     * Get all instructions in the transaction (method form).
     * Alias for the `instructions` property getter.
     * @returns Array of instructions with programId, accounts, and data
     */
    getInstructions() {
        return this.instructions;
    }
    /**
     * Add a signature for a given public key.
     *
     * The pubkey must be one of the required signers in the transaction.
     * The signature must be exactly 64 bytes (Ed25519 signature).
     *
     * @param pubkey - The public key as a base58 string
     * @param signature - The 64-byte signature as Uint8Array
     * @throws Error if pubkey is not a signer or signature is invalid
     *
     * @example
     * ```typescript
     * // Add a pre-computed signature (e.g., from TSS)
     * tx.addSignature(signerPubkey, signatureBytes);
     *
     * // Serialize and broadcast
     * const signedTxBytes = tx.toBytes();
     * ```
     */
    addSignature(pubkey, signature) {
        this._wasm.add_signature(pubkey, signature);
    }
    /**
     * Get the signer index for a public key.
     *
     * Returns the index in the signatures array where this pubkey's
     * signature should be placed, or null if the pubkey is not a signer.
     *
     * @param pubkey - The public key as a base58 string
     * @returns The signer index, or null if not a signer
     */
    signerIndex(pubkey) {
        const idx = this._wasm.signer_index(pubkey);
        return idx ?? null;
    }
    /**
     * Sign this transaction with a Keypair.
     *
     * Signs the transaction message and places the signature at the correct
     * signer index.
     *
     * @param keypair - A Keypair instance
     */
    signWithKeypair(keypair) {
        this._wasm.sign_with_keypair(keypair.wasm);
    }
    /**
     * Get the underlying WASM instance (internal use only)
     * @internal
     */
    get wasm() {
        return this._wasm;
    }
}
exports.Transaction = Transaction;
