// @bitgo/wasm-solana/dist/cjs/js/versioned.js
"use strict";
/**
 * Versioned Transaction Support
 *
 * Handles both legacy and versioned (MessageV0) Solana transactions.
 * Versioned transactions use Address Lookup Tables (ALTs) to compress
 * transaction size by referencing accounts via lookup table indices.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionedTransaction = void 0;
exports.isVersionedTransaction = isVersionedTransaction;
const wasm_solana_js_1 = require("./wasm/wasm_solana.js");
/**
 * Detect if transaction bytes represent a versioned transaction (MessageV0).
 *
 * @param bytes - Raw transaction bytes
 * @returns true if versioned, false if legacy
 */
function isVersionedTransaction(bytes) {
    return (0, wasm_solana_js_1.is_versioned_transaction)(bytes);
}
/**
 * Versioned Transaction class.
 *
 * Handles both legacy and versioned (MessageV0) transactions.
 * Provides access to Address Lookup Table data for versioned transactions.
 */
class VersionedTransaction {
    inner;
    constructor(inner) {
        this.inner = inner;
    }
    /**
     * Deserialize a transaction from raw bytes.
     * Automatically handles both legacy and versioned formats.
     */
    static fromBytes(bytes) {
        return new VersionedTransaction(wasm_solana_js_1.WasmVersionedTransaction.from_bytes(bytes));
    }
    /**
     * Create a VersionedTransaction from a WasmVersionedTransaction instance.
     * @internal Used by builder functions
     */
    static fromWasm(wasm) {
        return new VersionedTransaction(wasm);
    }
    /**
     * Create a versioned transaction from raw MessageV0 data.
     *
     * This is used for the `fromVersionedTransactionData()` path where we have
     * pre-compiled versioned data (indexes + ALT refs). No instruction compilation
     * is needed - this just constructs the transaction from the raw structure.
     *
     * @param data - Raw versioned transaction data
     * @returns A VersionedTransaction instance
     *
     * @example
     * ```typescript
     * const tx = VersionedTransaction.fromVersionedData({
     *   staticAccountKeys: ['pubkey1', 'pubkey2', ...],
     *   addressLookupTables: [
     *     { accountKey: 'altPubkey', writableIndexes: [0, 1], readonlyIndexes: [2] }
     *   ],
     *   versionedInstructions: [
     *     { programIdIndex: 0, accountKeyIndexes: [1, 2], data: 'base58EncodedData' }
     *   ],
     *   messageHeader: {
     *     numRequiredSignatures: 1,
     *     numReadonlySignedAccounts: 0,
     *     numReadonlyUnsignedAccounts: 3
     *   },
     *   recentBlockhash: 'blockhash'
     * });
     * ```
     */
    static fromVersionedData(data) {
        // Build the transaction using WASM and wrap in TypeScript class
        const wasm = wasm_solana_js_1.BuilderNamespace.build_from_versioned_data(data);
        return VersionedTransaction.fromWasm(wasm);
    }
    /**
     * Check if this is a versioned transaction (MessageV0).
     */
    get isVersioned() {
        return this.inner.is_versioned;
    }
    /**
     * Get the fee payer address.
     */
    get feePayer() {
        return this.inner.fee_payer ?? undefined;
    }
    /**
     * Get the recent blockhash.
     */
    get recentBlockhash() {
        return this.inner.recent_blockhash;
    }
    /**
     * Get the number of instructions.
     */
    get numInstructions() {
        return this.inner.num_instructions;
    }
    /**
     * Get the number of signatures.
     */
    get numSignatures() {
        return this.inner.num_signatures;
    }
    /**
     * Get the transaction ID (first signature as base58).
     *
     * For Solana, the transaction ID is the first signature.
     * Returns `undefined` if the transaction is unsigned (no signatures or all-zeros signature).
     */
    get id() {
        return this.inner.id;
    }
    /**
     * Get the signable message payload.
     */
    signablePayload() {
        return this.inner.signable_payload();
    }
    /**
     * Serialize the message portion of the transaction.
     * Alias for signablePayload() - provides compatibility with @solana/web3.js API.
     * @returns The serialized message bytes
     */
    serializeMessage() {
        return this.signablePayload();
    }
    /**
     * Serialize the transaction to bytes.
     */
    toBytes() {
        return this.inner.to_bytes();
    }
    /**
     * Serialize to network broadcast format.
     * @returns The transaction as bytes ready for broadcast
     */
    toBroadcastFormat() {
        return this.toBytes();
    }
    /**
     * Get static account keys (accounts stored directly in the message).
     * For versioned transactions, additional accounts may be referenced via ALTs.
     */
    staticAccountKeys() {
        return Array.from(this.inner.static_account_keys());
    }
    /**
     * Get Address Lookup Table data.
     * Returns empty array for legacy transactions.
     */
    addressLookupTables() {
        const alts = this.inner.address_lookup_tables();
        return Array.from(alts).map((alt) => ({
            accountKey: alt.accountKey,
            writableIndexes: alt.writableIndexes,
            readonlyIndexes: alt.readonlyIndexes,
        }));
    }
    /**
     * Get all signatures as byte arrays.
     * Provides compatibility with @solana/web3.js Transaction.signatures API.
     */
    get signatures() {
        return Array.from(this.inner.signatures());
    }
    /**
     * Add a signature for a given public key.
     *
     * @param pubkey - The public key as base58 string
     * @param signature - The 64-byte Ed25519 signature
     */
    addSignature(pubkey, signature) {
        this.inner.add_signature(pubkey, signature);
    }
    /**
     * Get the signer index for a public key.
     * Returns undefined if the pubkey is not a required signer.
     */
    signerIndex(pubkey) {
        return this.inner.signer_index(pubkey) ?? undefined;
    }
    /**
     * Get all instructions.
     * Note: For versioned transactions, account indices may reference
     * accounts from Address Lookup Tables.
     */
    instructions() {
        const instructions = this.inner.instructions();
        return Array.from(instructions);
    }
}
exports.VersionedTransaction = VersionedTransaction;
