// @bitgo/wasm-utxo/dist/esm/js/descriptorWallet/DescriptorOutput.js
/**
 * Descriptor output types and utilities.
 * Moved from @bitgo/utxo-core.
 */
import { Descriptor } from "../index.js";
import { getFixedOutputSum } from "./Output.js";
import { getDescriptorAtIndexCheckScript } from "./derive.js";
export function isInternalOutput(output) {
    return "descriptor" in output && output.descriptor !== undefined;
}
export function isExternalOutput(output) {
    return !isInternalOutput(output);
}
/**
 * @return the sum of the external outputs that are not 'max'
 * @param outputs
 */
export function getExternalFixedAmount(outputs) {
    return getFixedOutputSum(outputs.filter(isExternalOutput));
}
export function toDerivedDescriptorWalletOutput(output, descriptorMap) {
    const descriptor = descriptorMap.get(output.descriptorName);
    if (!descriptor) {
        throw new Error(`Descriptor not found: ${output.descriptorName}`);
    }
    if (!(descriptor instanceof Descriptor)) {
        throw new Error(`Expected Descriptor instance for ${output.descriptorName}`);
    }
    const descriptorAtIndex = getDescriptorAtIndexCheckScript(descriptor, output.descriptorIndex, output.witnessUtxo.script, output.descriptorName);
    return {
        hash: output.hash,
        index: output.index,
        witnessUtxo: output.witnessUtxo,
        descriptor: descriptorAtIndex,
    };
}
