// @bitgo/sdk-coin-canton/dist/test/integration/canton.token.integration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const resources_1 = require("../resources");
const src_1 = require("../../src");
describe('Canton Token integration tests', function () {
    const tokenName = 'tcanton:testtoken';
    let bitgo;
    let basecoin;
    let newTxPrebuild;
    let newTxParams;
    let wallet;
    const txPrebuild = {
        txHex: resources_1.CantonTokenTransferRawTxn,
        txInfo: {},
    };
    const txParams = {
        recipients: [
            {
                address: resources_1.TokenTxParams.RECIPIENT_ADDRESS,
                amount: resources_1.TokenTxParams.AMOUNT,
                tokenName: resources_1.TokenTxParams.TOKEN,
            },
        ],
    };
    before(() => {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        src_1.CantonToken.createTokenConstructors().forEach(({ name, coinConstructor }) => {
            bitgo.safeRegister(name, coinConstructor);
        });
        basecoin = bitgo.coin(tokenName);
        newTxPrebuild = () => {
            return structuredClone(txPrebuild);
        };
        newTxParams = () => {
            return structuredClone(txParams);
        };
        wallet = new sdk_core_1.Wallet(bitgo, basecoin, {});
    });
    describe('Verify Transaction', function () {
        it('should verify token transfer transaction', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            const isTxnVerifies = await basecoin.verifyTransaction({ txPrebuild: txPrebuild, txParams: txParams, wallet });
            isTxnVerifies.should.equal(true);
        });
    });
});
