// @bitgo/sdk-coin-coredao/dist/test/unit/coredao.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should = __importStar(require("should"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const src_1 = require("../../src");
const nock_1 = __importDefault(require("nock"));
const tx_1 = require("@ethereumjs/tx");
const util_1 = require("@ethereumjs/util");
const sdk_core_1 = require("@bitgo/sdk-core");
const resources_1 = require("../resources");
const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
describe('Coredao', function () {
    before(function () {
        bitgo.safeRegister('coredao', src_1.Coredao.createInstance);
        bitgo.safeRegister('tcoredao', src_1.Tcoredao.createInstance);
        bitgo.initializeTestVars();
    });
    describe('Basic Coin Info', function () {
        it('should return the right info for coredao', function () {
            const coredao = bitgo.coin('coredao');
            coredao.should.be.an.instanceof(src_1.Coredao);
            coredao.getChain().should.equal('coredao');
            coredao.getFamily().should.equal('coredao');
            coredao.getFullName().should.equal('Core');
            coredao.getBaseFactor().should.equal(1e18);
            coredao.supportsTss().should.equal(true);
            coredao.allowsAccountConsolidations().should.equal(false);
        });
        it('should return the right info for tcoredao', function () {
            const tcoredao = bitgo.coin('tcoredao');
            tcoredao.should.be.an.instanceof(src_1.Tcoredao);
            tcoredao.getChain().should.equal('tcoredao');
            tcoredao.getFamily().should.equal('coredao');
            tcoredao.getFullName().should.equal('Testnet Core');
            tcoredao.getBaseFactor().should.equal(1e18);
            tcoredao.supportsTss().should.equal(true);
            tcoredao.allowsAccountConsolidations().should.equal(false);
        });
    });
    describe('Build Recover Transaction', function () {
        const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
        const explorerUrl = sdk_core_1.common.Environments[bitgo.getEnv()].coredaoExplorerBaseUrl;
        it('should generate a signed non-bitgo recovery tx', async () => {
            (0, nock_1.default)(explorerUrl)
                .get('/api')
                .twice()
                .query(resources_1.mockDataNonBitGoRecovery.getTxListRequest)
                .reply(200, resources_1.mockDataNonBitGoRecovery.getTxListResponse);
            (0, nock_1.default)(explorerUrl)
                .get('/api')
                .query(resources_1.mockDataNonBitGoRecovery.getBalanceRequest)
                .reply(200, resources_1.mockDataNonBitGoRecovery.getBalanceResponse);
            const baseCoin = bitgo.coin('tcoredao');
            const transaction = await baseCoin.recover({
                userKey: resources_1.mockDataNonBitGoRecovery.userKeyData,
                backupKey: resources_1.mockDataNonBitGoRecovery.backupKeyData,
                walletContractAddress: resources_1.mockDataNonBitGoRecovery.walletRootAddress,
                walletPassphrase: resources_1.mockDataNonBitGoRecovery.walletPassphrase,
                recoveryDestination: resources_1.mockDataNonBitGoRecovery.recoveryDestination,
                isTss: true,
                eip1559: { maxFeePerGas: 20000000000, maxPriorityFeePerGas: 10000000000 },
                gasLimit: 500000,
                replayProtectionOptions: {
                    chain: 1114,
                    hardfork: 'london',
                },
            });
            should.exist(transaction);
            transaction.should.have.property('id');
            transaction.should.have.property('tx');
            const tx = tx_1.FeeMarketEIP1559Transaction.fromSerializedTx(Buffer.from((0, util_1.stripHexPrefix)(transaction.tx), 'hex'));
            tx.getSenderAddress().toString().should.equal(resources_1.mockDataNonBitGoRecovery.walletRootAddress);
            const jsonTx = tx.toJSON();
            jsonTx.chainId?.should.equal('0x45a');
            jsonTx.to?.should.equal(resources_1.mockDataNonBitGoRecovery.recoveryDestination);
        });
    });
});
