// @bitgo/sdk-coin-iota/dist/test/unit/transactionBuilder/transactionBuilderFactory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const statics_1 = require("@bitgo/statics");
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/iota"));
const testHelpers_1 = require("../helpers/testHelpers");
describe('Iota Transaction Builder Factory', () => {
    const factory = new src_1.TransactionBuilderFactory(statics_1.coins.get('tiota'));
    describe('getTransferBuilder', () => {
        it('should create a transfer builder', function () {
            const builder = factory.getTransferBuilder();
            should_1.default.exist(builder);
            (0, should_1.default)(builder instanceof src_1.TransferBuilder).be.true();
        });
        it('should create a transfer builder with existing transaction', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx = (await txBuilder.build());
            const newBuilder = factory.getTransferBuilder(tx);
            should_1.default.exist(newBuilder);
            (0, should_1.default)(newBuilder instanceof src_1.TransferBuilder).be.true();
            const rebuiltTx = (await newBuilder.build());
            should_1.default.equal(rebuiltTx.sender, tx.sender);
            should_1.default.deepEqual(rebuiltTx.recipients, tx.recipients);
        });
    });
    describe('from', () => {
        const formats = [
            {
                name: 'hex string',
                convert: (rawTx) => Buffer.from(rawTx, 'base64').toString('hex'),
            },
            {
                name: 'Uint8Array',
                convert: (rawTx) => Buffer.from(rawTx, 'base64'),
            },
        ];
        formats.forEach(({ name, convert }) => {
            it(`should create builder from raw transaction in ${name} format`, async function () {
                const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
                const tx = (await txBuilder.build());
                const rawTx = await tx.toBroadcastFormat();
                const convertedTx = convert(rawTx);
                const rebuiltBuilder = factory.from(convertedTx);
                should_1.default.exist(rebuiltBuilder);
                (0, should_1.default)(rebuiltBuilder instanceof src_1.TransferBuilder).be.true();
                const rebuiltTx = (await rebuiltBuilder.build());
                rebuiltTx.sender.should.equal(testData.sender.address);
                rebuiltTx.recipients.length.should.equal(testData.recipients.length);
                rebuiltTx.recipients[0].address.should.equal(testData.recipients[0].address);
                rebuiltTx.recipients[0].amount.should.equal(testData.recipients[0].amount);
                should_1.default.exist(rebuiltTx.paymentObjects);
                rebuiltTx.paymentObjects?.length.should.equal(testData.paymentObjects.length);
            });
        });
    });
    describe('Error Handling', () => {
        it('should throw error for invalid raw transaction', function () {
            (0, should_1.default)(() => factory.from('invalid-raw-transaction')).throwError();
        });
        it('should throw error for empty raw transaction', function () {
            (0, should_1.default)(() => factory.from('')).throwError();
        });
        it('should throw error for unsupported transaction type', async function () {
            // This would require creating a transaction with unsupported commands
            // For now, we test that invalid data throws an error
            const invalidData = Buffer.from('not-a-valid-transaction').toString('base64');
            (0, should_1.default)(() => factory.from(invalidData)).throwError();
        });
        it('should throw error when getWalletInitializationBuilder is called', function () {
            (0, should_1.default)(() => factory.getWalletInitializationBuilder()).throwError('Method not implemented.');
        });
    });
    describe('Transaction Rebuilding and Modification', () => {
        it('should allow creating transactions with modified gas budget', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx = (await txBuilder.build());
            should_1.default.equal(tx.gasBudget, testData.GAS_BUDGET);
            const modifiedBuilder = factory.getTransferBuilder();
            modifiedBuilder.sender(testData.sender.address);
            modifiedBuilder.recipients(testData.recipients);
            modifiedBuilder.paymentObjects(testData.paymentObjects);
            modifiedBuilder.gasData({
                gasBudget: testData.GAS_BUDGET * 2,
                gasPrice: testData.GAS_PRICE,
                gasPaymentObjects: testData.gasPaymentObjects,
            });
            const modifiedTx = (await modifiedBuilder.build());
            should_1.default.equal(modifiedTx.gasBudget, testData.GAS_BUDGET * 2);
        });
        it('should build transaction with gas sponsor', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            txBuilder.gasSponsor(testData.gasSponsor.address);
            const tx = (await txBuilder.build());
            should_1.default.equal(tx.sender, testData.sender.address);
            should_1.default.equal(tx.gasSponsor, testData.gasSponsor.address);
            should_1.default.notEqual(tx.sender, tx.gasSponsor);
        });
        it('should maintain transaction ID for identical transactions', async function () {
            const txBuilder1 = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx1 = (await txBuilder1.build());
            const txBuilder2 = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx2 = (await txBuilder2.build());
            should_1.default.equal(tx1.id, tx2.id);
        });
    });
    describe('Builder Initialization', () => {
        it('should initialize builder with transaction correctly', async function () {
            const originalBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            const originalTx = (await originalBuilder.build());
            const newBuilder = factory.getTransferBuilder(originalTx);
            const newTx = (await newBuilder.build());
            should_1.default.equal(newTx.sender, originalTx.sender);
            should_1.default.deepEqual(newTx.recipients, originalTx.recipients);
            should_1.default.equal(newTx.gasBudget, originalTx.gasBudget);
            should_1.default.equal(newTx.gasPrice, originalTx.gasPrice);
        });
    });
    describe('Round Trip Conversion', () => {
        it('should handle JSON round trip', async function () {
            const originalBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            const originalTx = (await originalBuilder.build());
            const json = originalTx.toJson();
            const newTx = new src_1.TransferTransaction(statics_1.coins.get('tiota'));
            newTx.parseFromJSON(json);
            const newBuilder = factory.getTransferBuilder(newTx);
            const rebuiltTx = (await newBuilder.build());
            should_1.default.equal(rebuiltTx.sender, originalTx.sender);
            should_1.default.deepEqual(rebuiltTx.recipients, originalTx.recipients);
        });
        it('should serialize to consistent format', async function () {
            const builder1 = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx1 = (await builder1.build());
            const serialized1 = await tx1.toBroadcastFormat();
            const builder2 = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx2 = (await builder2.build());
            const serialized2 = await tx2.toBroadcastFormat();
            should_1.default.equal(serialized1, serialized2);
        });
    });
});
