// @bitgo/sdk-coin-iota/dist/test/unit/transactionBuilder/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../../../src/lib/constants");
const testHelpers_1 = require("../helpers/testHelpers");
describe('Iota Transfer Builder', () => {
    const factory = (0, testHelpers_1.getFactory)();
    describe('Basic Transaction Building', () => {
        it('should build a transfer transaction in simulate mode', async function () {
            const txBuilder = (0, testHelpers_1.createBasicTransferBuilder)();
            const tx = (await txBuilder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.isSimulateTx, true);
            (0, testHelpers_1.assertBasicTransactionProperties)(tx);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should build a transfer transaction with gas data', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx = (await txBuilder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.isSimulateTx, false);
            (0, testHelpers_1.assertBasicTransactionProperties)(tx);
            (0, testHelpers_1.assertGasData)(tx);
            (0, testHelpers_1.assertInputsAndOutputs)(tx, testHelpers_1.testData.recipients);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should build a transfer transaction with gas sponsor', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithSponsor)();
            const tx = (await txBuilder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.sender, testHelpers_1.testData.sender.address);
            should_1.default.equal(tx.gasSponsor, testHelpers_1.testData.gasSponsor.address);
            should_1.default.equal(tx.isSimulateTx, false);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should build transaction using gas objects when no payment objects', async function () {
            const txBuilder = (0, testHelpers_1.createBuilderWithGasObjectsOnly)();
            const tx = (await txBuilder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.isSimulateTx, false);
            should_1.default.equal(tx.paymentObjects, undefined);
            should_1.default.exist(tx.gasPaymentObjects);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should support method chaining', async function () {
            const tx = (await factory
                .getTransferBuilder()
                .sender(testHelpers_1.testData.sender.address)
                .recipients(testHelpers_1.testData.recipients)
                .paymentObjects(testHelpers_1.testData.paymentObjects)
                .gasData(testHelpers_1.testData.gasData)
                .gasSponsor(testHelpers_1.testData.gasSponsor.address)
                .build());
            should_1.default.exist(tx);
            should_1.default.equal(tx.sender, testHelpers_1.testData.sender.address);
            should_1.default.equal(tx.gasSponsor, testHelpers_1.testData.gasSponsor.address);
        });
    });
    describe('Serialization and Deserialization', () => {
        it('should serialize to JSON and rebuild transaction', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            const tx = (await txBuilder.build());
            const txJson = tx.toJson();
            const rebuiltTxBuilder = factory.getTransferBuilder();
            rebuiltTxBuilder.transaction.parseFromJSON(txJson);
            const rebuiltTx = (await rebuiltTxBuilder.build());
            should_1.default.equal(rebuiltTx.sender, tx.sender);
            should_1.default.deepEqual(rebuiltTx.recipients, tx.recipients);
            should_1.default.equal(rebuiltTx.gasBudget, tx.gasBudget);
            should_1.default.equal(rebuiltTx.gasPrice, tx.gasPrice);
        });
        it('should serialize to broadcast format consistently', async function () {
            const tx1 = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            const rawTx1 = await tx1.toBroadcastFormat();
            const tx2 = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            const rawTx2 = await tx2.toBroadcastFormat();
            should_1.default.equal(rawTx1, rawTx2);
        });
        it('should validate toJSON output', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            const txJson = tx.toJson();
            should_1.default.equal(txJson.sender, testHelpers_1.testData.sender.address);
            should_1.default.deepEqual(txJson.recipients, testHelpers_1.testData.recipients);
            should_1.default.deepEqual(txJson.paymentObjects, testHelpers_1.testData.paymentObjects);
            should_1.default.equal(txJson.gasBudget, testHelpers_1.testData.GAS_BUDGET);
            should_1.default.equal(txJson.gasPrice, testHelpers_1.testData.GAS_PRICE);
            should_1.default.deepEqual(txJson.gasPaymentObjects, testHelpers_1.testData.gasPaymentObjects);
            should_1.default.equal(txJson.type, sdk_core_1.TransactionType.Send);
        });
    });
    describe('Round-trip Parsing', () => {
        it('should correctly parse and rebuild transaction using gas objects', async function () {
            const tx = (await (0, testHelpers_1.createBuilderWithGasObjectsOnly)().build());
            const rawTx = await tx.toBroadcastFormat();
            const txHex = Buffer.from(rawTx, 'base64').toString('hex');
            const rebuiltBuilder = factory.from(txHex);
            const rebuiltTx = (await rebuiltBuilder.build());
            should_1.default.equal(rebuiltTx.sender, tx.sender);
            should_1.default.deepEqual(rebuiltTx.recipients, tx.recipients);
            should_1.default.equal(rebuiltTx.gasBudget, tx.gasBudget);
            should_1.default.equal(rebuiltTx.gasPrice, tx.gasPrice);
            should_1.default.equal(rebuiltTx.paymentObjects, undefined);
            should_1.default.exist(rebuiltTx.gasPaymentObjects);
        });
        it('should correctly parse transaction with payment objects', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            const rawTx = await tx.toBroadcastFormat();
            const txHex = Buffer.from(rawTx, 'base64').toString('hex');
            const rebuiltBuilder = factory.from(txHex);
            const rebuiltTx = (await rebuiltBuilder.build());
            should_1.default.exist(rebuiltTx.paymentObjects);
            should_1.default.equal(rebuiltTx.paymentObjects?.length, testHelpers_1.testData.paymentObjects.length);
        });
        it('should handle round-trip with gas sponsor and payment objects', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithSponsor)().build());
            const rawTx = await tx.toBroadcastFormat();
            const txHex = Buffer.from(rawTx, 'base64').toString('hex');
            const rebuiltBuilder = factory.from(txHex);
            const rebuiltTx = (await rebuiltBuilder.build());
            should_1.default.equal(rebuiltTx.gasSponsor, testHelpers_1.testData.gasSponsor.address);
            should_1.default.exist(rebuiltTx.paymentObjects);
        });
    });
    describe('Recipient Handling', () => {
        const testCases = [
            { count: 1, description: 'single recipient' },
            { count: 10, description: 'multiple recipients' },
            { count: constants_1.MAX_RECIPIENTS, description: 'maximum recipients' },
        ];
        testCases.forEach(({ count, description }) => {
            it(`should build transaction with ${description}`, async function () {
                const recipients = (0, testHelpers_1.createRecipients)(count, '1000');
                const tx = (await (0, testHelpers_1.createBuilderWithRecipients)(recipients).build());
                should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
                tx.outputs.length.should.equal(count);
                tx.outputs.forEach((output) => {
                    output.value.should.equal('1000');
                    output.coin?.should.equal('tiota');
                });
            });
        });
        it('should fail with more than MAX_RECIPIENTS', async function () {
            const recipients = (0, testHelpers_1.createRecipients)(constants_1.MAX_RECIPIENTS + 1);
            const builder = (0, testHelpers_1.createBuilderWithRecipients)(recipients);
            await builder.build().should.be.rejected();
        });
        it('should handle duplicate recipient addresses', async function () {
            const duplicateRecipients = [
                { address: testHelpers_1.testData.addresses.validAddresses[0], amount: '1000' },
                { address: testHelpers_1.testData.addresses.validAddresses[0], amount: '2000' },
            ];
            const tx = (await (0, testHelpers_1.createBuilderWithRecipients)(duplicateRecipients).build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.outputs.length.should.equal(2);
        });
    });
    describe('Gas Object Handling', () => {
        it('should build with single gas object', async function () {
            const tx = (await (0, testHelpers_1.createBuilderWithGasObjects)([testHelpers_1.testData.gasPaymentObjects[0]]).build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.exist(tx.gasPaymentObjects);
            tx.gasPaymentObjects?.length.should.equal(1);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should build with exactly MAX_GAS_PAYMENT_OBJECTS gas objects', async function () {
            // Create exactly MAX_GAS_PAYMENT_OBJECTS (256) with valid digests
            const exactlyMaxGasObjects = [];
            for (let i = 0; i < constants_1.MAX_GAS_PAYMENT_OBJECTS; i++) {
                exactlyMaxGasObjects.push({
                    ...testHelpers_1.testData.gasPaymentObjects[i % testHelpers_1.testData.gasPaymentObjects.length],
                    objectId: `0x${i.toString(16).padStart(64, '0')}`,
                });
            }
            const tx = (await (0, testHelpers_1.createBuilderWithGasObjects)(exactlyMaxGasObjects).build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.sender, testHelpers_1.testData.sender.address);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should build with more than MAX_GAS_PAYMENT_OBJECTS requiring merge', async function () {
            // Create more than MAX_GAS_PAYMENT_OBJECTS (256) to test merge logic
            const manyGasObjects = [];
            for (let i = 0; i < constants_1.MAX_GAS_PAYMENT_OBJECTS + 10; i++) {
                manyGasObjects.push({
                    ...testHelpers_1.testData.gasPaymentObjects[i % testHelpers_1.testData.gasPaymentObjects.length],
                    objectId: `0x${i.toString(16).padStart(64, '0')}`,
                });
            }
            const tx = (await (0, testHelpers_1.createBuilderWithGasObjects)(manyGasObjects).build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.isSimulateTx, false);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
    });
    describe('Payment Object Handling', () => {
        it('should build with single payment object', async function () {
            const builder = factory
                .getTransferBuilder()
                .sender(testHelpers_1.testData.sender.address)
                .recipients(testHelpers_1.testData.recipients)
                .paymentObjects([testHelpers_1.testData.paymentObjects[0]])
                .gasData(testHelpers_1.testData.gasData);
            const tx = (await builder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.paymentObjects?.length.should.equal(1);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should keep payment and gas objects separate', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            should_1.default.exist(tx.paymentObjects);
            should_1.default.exist(tx.gasPaymentObjects);
            tx.paymentObjects?.length.should.equal(testHelpers_1.testData.paymentObjects.length);
            tx.gasPaymentObjects?.length.should.equal(testHelpers_1.testData.gasPaymentObjects.length);
        });
        it('should handle payment objects with different versions', async function () {
            const mixedVersionObjects = [
                {
                    objectId: '0x1111111111111111111111111111111111111111111111111111111111111111',
                    version: '1',
                    digest: 'DGVhYjk6YHwdPdZBgBN8czavy8LvbrshkbxF963EW7mB',
                },
                {
                    objectId: '0x2222222222222222222222222222222222222222222222222222222222222222',
                    version: '999999',
                    digest: 'DoJwXuz9oU5Y5v5vBRiTgisVTQuZQLmHZWeqJzzD5QUE',
                },
            ];
            const builder = factory
                .getTransferBuilder()
                .sender(testHelpers_1.testData.sender.address)
                .recipients(testHelpers_1.testData.recipients)
                .paymentObjects(mixedVersionObjects)
                .gasData(testHelpers_1.testData.gasData);
            const tx = (await builder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.paymentObjects?.length.should.equal(2);
        });
        it('should accept empty payment objects array and use gas objects for payment in non-sponsor mode', async function () {
            const builder = factory
                .getTransferBuilder()
                .sender(testHelpers_1.testData.sender.address)
                .recipients(testHelpers_1.testData.recipients)
                .paymentObjects([])
                .gasData(testHelpers_1.testData.gasData);
            const tx = (await builder.build());
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            should_1.default.equal(tx.isSimulateTx, false);
            should_1.default.equal(tx.sender, testHelpers_1.testData.sender.address);
            await (0, testHelpers_1.assertValidRawTransaction)(tx);
        });
        it('should fail tx building in case of empty payment objects in sponsor mode', async function () {
            const builder = factory
                .getTransferBuilder()
                .sender(testHelpers_1.testData.sender.address)
                .gasSponsor(testHelpers_1.testData.gasSponsor.address)
                .recipients(testHelpers_1.testData.recipients)
                .paymentObjects([])
                .gasData(testHelpers_1.testData.gasData);
            await builder.build().should.be.rejectedWith(/Payment objects are required when using a gas sponsor/);
        });
    });
    describe('Validation Errors', () => {
        describe('Address Validation', () => {
            it('should fail for invalid sender', function () {
                const builder = factory.getTransferBuilder();
                (0, should_1.default)(() => builder.sender('randomString')).throwError('Invalid address randomString');
            });
            it('should fail for invalid recipient address', function () {
                const builder = (0, testHelpers_1.createBasicTransferBuilder)();
                (0, should_1.default)(() => builder.recipients([{ address: 'invalidAddress', amount: '1000' }])).throwError('Invalid address invalidAddress');
            });
        });
        describe('Amount Validation', () => {
            it('should fail for negative amount', function () {
                const builder = (0, testHelpers_1.createBasicTransferBuilder)();
                (0, should_1.default)(() => builder.recipients([{ address: testHelpers_1.testData.addresses.validAddresses[0], amount: '-1000' }])).throwError('Value cannot be less than zero');
            });
            it('should fail for invalid format', function () {
                const builder = (0, testHelpers_1.createBasicTransferBuilder)();
                (0, should_1.default)(() => builder.recipients([{ address: testHelpers_1.testData.addresses.validAddresses[0], amount: 'invalid' }])).throw();
            });
            it('should accept zero amount', async function () {
                const recipients = [{ address: testHelpers_1.testData.addresses.validAddresses[0], amount: '0' }];
                const tx = (await (0, testHelpers_1.createBuilderWithRecipients)(recipients).build());
                should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
                tx.outputs[0].value.should.equal('0');
            });
            it('should accept very large amount', async function () {
                const largeAmount = '999999999999999999';
                const recipients = [{ address: testHelpers_1.testData.addresses.validAddresses[0], amount: largeAmount }];
                const tx = (await (0, testHelpers_1.createBuilderWithRecipients)(recipients).build());
                should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
                tx.outputs[0].value.should.equal(largeAmount);
            });
        });
        describe('Required Field Validation', () => {
            it('should fail when missing sender', async function () {
                const builder = factory
                    .getTransferBuilder()
                    .recipients(testHelpers_1.testData.recipients)
                    .paymentObjects(testHelpers_1.testData.paymentObjects)
                    .gasData(testHelpers_1.testData.gasData);
                await builder.build().should.be.rejected();
            });
            it('should fail when missing recipients', async function () {
                const builder = factory
                    .getTransferBuilder()
                    .sender(testHelpers_1.testData.sender.address)
                    .paymentObjects(testHelpers_1.testData.paymentObjects)
                    .gasData(testHelpers_1.testData.gasData);
                await builder.build().should.be.rejected();
            });
            it('should fail for empty recipients', async function () {
                const builder = factory
                    .getTransferBuilder()
                    .sender(testHelpers_1.testData.sender.address)
                    .recipients([])
                    .paymentObjects(testHelpers_1.testData.paymentObjects)
                    .gasData(testHelpers_1.testData.gasData);
                await builder.build().should.be.rejected();
            });
            it('should allow empty payment objects', function () {
                const builder = (0, testHelpers_1.createBasicTransferBuilder)();
                should_1.default.doesNotThrow(() => builder.paymentObjects([]));
            });
            it('should fail without payment objects when using gas sponsor', async function () {
                const builder = factory
                    .getTransferBuilder()
                    .sender(testHelpers_1.testData.sender.address)
                    .recipients(testHelpers_1.testData.recipients)
                    .gasData(testHelpers_1.testData.gasData)
                    .gasSponsor(testHelpers_1.testData.gasSponsor.address);
                await builder.build().should.be.rejected();
            });
        });
        describe('Gas Data Validation', () => {
            it('should fail for empty gas payment objects', function () {
                const builder = (0, testHelpers_1.createBasicTransferBuilder)();
                (0, should_1.default)(() => builder.gasData({
                    gasBudget: testHelpers_1.testData.GAS_BUDGET,
                    gasPrice: testHelpers_1.testData.GAS_PRICE,
                    gasPaymentObjects: [],
                })).throwError(/Gas input objects list is empty/);
            });
        });
        describe('Object Duplication Validation', () => {
            it('should fail when payment and gas payment objects overlap', async function () {
                const builder = factory
                    .getTransferBuilder()
                    .sender(testHelpers_1.testData.sender.address)
                    .recipients(testHelpers_1.testData.recipients)
                    .paymentObjects(testHelpers_1.testData.gasPaymentObjects)
                    .gasData(testHelpers_1.testData.gasData);
                await builder.build().should.be.rejected();
            });
        });
        describe('Transaction Parsing Validation', () => {
            it('should fail to parse invalid raw transaction', function () {
                (0, should_1.default)(() => factory.from('invalidRawTransaction')).throwError();
            });
        });
    });
    describe('Transaction Signing', () => {
        it('should build transaction with sender signature', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            tx.addSignature(testHelpers_1.testData.testSignature.publicKey, testHelpers_1.testData.testSignature.signature);
            await tx.build();
            should_1.default.exist(tx.serializedSignature);
            should_1.default.equal(tx.serializedSignature.length > 0, true);
        });
        it('should build transaction with gas sponsor signature', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithSponsor)().build());
            tx.addGasSponsorSignature(testHelpers_1.testData.testGasSponsorSignature.publicKey, testHelpers_1.testData.testGasSponsorSignature.signature);
            await tx.build();
            should_1.default.exist(tx.serializedGasSponsorSignature);
            should_1.default.equal(tx.serializedGasSponsorSignature.length > 0, true);
        });
        it('should build transaction with both sender and gas sponsor signatures', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithSponsor)().build());
            tx.addSignature(testHelpers_1.testData.testSignature.publicKey, testHelpers_1.testData.testSignature.signature);
            tx.addGasSponsorSignature(testHelpers_1.testData.testGasSponsorSignature.publicKey, testHelpers_1.testData.testGasSponsorSignature.signature);
            await tx.build();
            should_1.default.exist(tx.serializedSignature);
            should_1.default.exist(tx.serializedGasSponsorSignature);
            tx.signature.length.should.equal(2);
            tx.signature[0].should.equal(tx.serializedSignature);
            tx.signature[1].should.equal(tx.serializedGasSponsorSignature);
        });
        it('should add signature through builder and serialize correctly', async function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            txBuilder.addSignature(testHelpers_1.testData.testSignature.publicKey, testHelpers_1.testData.testSignature.signature);
            const tx = (await txBuilder.build());
            should_1.default.exist(tx.serializedSignature);
            should_1.default.equal(typeof tx.serializedSignature, 'string');
            tx.signature.length.should.equal(1);
        });
        it('should fail to add invalid sender signature via builder', function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithGas)();
            (0, should_1.default)(() => txBuilder.addSignature({ pub: 'tooshort' }, testHelpers_1.testData.testSignature.signature)).throwError('Invalid transaction signature');
        });
        it('should fail to add invalid gas sponsor signature via builder', function () {
            const txBuilder = (0, testHelpers_1.createTransferBuilderWithSponsor)();
            (0, should_1.default)(() => txBuilder.addGasSponsorSignature(testHelpers_1.testData.testGasSponsorSignature.publicKey, Buffer.from('invalid'))).throwError('Invalid transaction signature');
        });
    });
    describe('Signable Payload', () => {
        it('should generate signable payload for non-simulate transaction', async function () {
            const tx = (await (0, testHelpers_1.createTransferBuilderWithGas)().build());
            should_1.default.equal(tx.isSimulateTx, false);
            const signablePayload = tx.signablePayload;
            should_1.default.exist(signablePayload);
            should_1.default.equal(Buffer.isBuffer(signablePayload), true);
            should_1.default.equal(signablePayload.length, 32); // Blake2b hash is 32 bytes
        });
        it('should fail to get signable payload for simulate transaction', async function () {
            const tx = (await (0, testHelpers_1.createBasicTransferBuilder)().build());
            should_1.default.equal(tx.isSimulateTx, true);
            (0, should_1.default)(() => tx.signablePayload).throwError('Cannot sign a simulate tx');
        });
    });
    describe('Input and Output Calculation', () => {
        it('should calculate total input amount correctly for multiple recipients', async function () {
            const recipients = [
                { address: testHelpers_1.testData.addresses.validAddresses[0], amount: '1000' },
                { address: testHelpers_1.testData.addresses.validAddresses[1], amount: '2000' },
                { address: testHelpers_1.testData.addresses.validAddresses[2], amount: '3000' },
            ];
            const tx = (await (0, testHelpers_1.createBuilderWithRecipients)(recipients).build());
            const expectedTotal = '6000';
            tx.inputs[0].value.should.equal(expectedTotal);
            tx.outputs.length.should.equal(3);
        });
    });
});
