// @bitgo/sdk-coin-near/dist/src/nep141Token.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Nep141Token = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
const near_1 = require("./near");
class Nep141Token extends near_1.Near {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === statics_1.NetworkType.MAINNET ? statics_1.coins.get('near') : statics_1.coins.get('tnear');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new Nep141Token(bitgo, config);
    }
    static createTokenConstructors(tokenConfig = [...statics_1.tokens.bitcoin.near.tokens, ...statics_1.tokens.testnet.near.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfig) {
            const tokenConstructor = Nep141Token.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get contractAddress() {
        return this.tokenConfig.contractAddress;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    get storageDepositAmount() {
        return this.tokenConfig.storageDepositAmount;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Nep141 Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    async verifyTransaction(params) {
        const { txPrebuild: txPrebuild, txParams: txParams, wallet } = params;
        const rawTx = txPrebuild.txHex;
        let totalAmount = new bignumber_js_1.default(0);
        if (!rawTx) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const coinConfig = statics_1.coins.get(this.getChain());
        const transaction = new lib_1.Transaction(coinConfig);
        transaction.fromRawTransaction(rawTx);
        const explainedTx = transaction.explainTransaction();
        if (txParams.recipients !== undefined) {
            txParams.recipients.forEach((recipient) => {
                if (recipient.tokenName && recipient.tokenName !== coinConfig.name) {
                    throw new Error('incorrect token name specified in recipients');
                }
                recipient.tokenName = coinConfig.name;
            });
            const filteredRecipients = txParams.recipients?.map((recipient) => ({
                address: recipient.address,
                amount: recipient.amount,
                tokenName: recipient.tokenName,
            }));
            const filteredOutputs = explainedTx.outputs.map((output) => ({
                address: output.address,
                amount: output.amount,
                tokenName: output.tokenName,
            }));
            const outputsMatch = JSON.stringify(filteredRecipients) === JSON.stringify(filteredOutputs);
            if (!outputsMatch) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            for (const recipient of txParams.recipients) {
                totalAmount = totalAmount.plus(recipient.amount);
            }
            if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                throw new Error('Tx total amount does not match with expected total amount field');
            }
        }
        if (params.verification?.consolidationToBaseAddress) {
            if (!wallet?.coinSpecific()?.rootAddress) {
                throw new Error('Unable to determine base address for consolidation');
            }
            await this.verifyConsolidationToBaseAddress(explainedTx, wallet.coinSpecific()?.rootAddress);
        }
        return true;
    }
}
exports.Nep141Token = Nep141Token;
