// @bitgo/sdk-core/dist/src/account-lib/baseCoin/messages/baseMessageBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseMessageBuilder = void 0;
const bitgo_1 = require("../../../bitgo");
const iface_1 = require("../iface");
/**
 * Base Message Builder
 */
class BaseMessageBuilder {
    /**
     * Base constructor.
     * @param coinConfig BaseCoin from statics library
     * @param messageType The type of message this builder creates, defaults to UNKNOWN
     */
    constructor(coinConfig, messageType = bitgo_1.MessageStandardType.UNKNOWN) {
        this.payload = '';
        this.signatures = [];
        this.signers = [];
        this.metadata = {};
        this.coinConfig = coinConfig;
        this.type = messageType;
    }
    /**
     * Sets the message payload to be used when building the message
     * @param payload The message payload (string, JSON, etc.)
     * @returns The builder instance for chaining
     */
    setPayload(payload) {
        this.payload = payload;
        return this;
    }
    /**
     * Sets metadata for the message
     * @param metadata Additional metadata for the message
     * @returns The builder instance for chaining
     */
    setMetadata(metadata) {
        this.metadata = metadata;
        return this;
    }
    /**
     * Gets the current message payload
     * @returns The current message payload
     */
    getPayload() {
        return this.payload;
    }
    /**
     * Gets the current metadata
     * @returns The current metadata
     */
    getMetadata() {
        return this.metadata;
    }
    getType() {
        return this.type;
    }
    setType(value) {
        this.type = value;
        return this;
    }
    getSignatures() {
        return this.signatures;
    }
    setSignatures(signatures) {
        this.signatures = signatures;
        return this;
    }
    addSignature(signature) {
        if (!this.signatures.some((sig) => sig.publicKey.pub === signature.publicKey.pub && sig.signature.equals(signature.signature))) {
            this.signatures.push(signature);
        }
        return this;
    }
    getSigners() {
        return this.signers;
    }
    setSigners(value) {
        this.signers = value;
        return this;
    }
    addSigner(signer) {
        if (!this.signers.includes(signer)) {
            this.signers.push(signer);
        }
        return this;
    }
    getDigest() {
        return this.digest;
    }
    setDigest(value) {
        this.digest = value;
        return this;
    }
    /**
     * Builds a message using the previously set payload and metadata
     * @returns A Promise resolving to the built IMessage
     */
    async build() {
        try {
            if (!this.payload) {
                throw new Error('Message payload must be set before building the message');
            }
            return this.buildMessage({
                coinConfig: this.coinConfig,
                payload: this.payload,
                type: this.type,
                signatures: this.signatures,
                signers: this.signers,
                metadata: {
                    ...this.metadata,
                    encoding: 'utf8',
                },
            });
        }
        catch (err) {
            if (err instanceof Error) {
                throw err;
            }
            throw new Error(`Failed to build message of type ${this.type}`);
        }
    }
    /**
     * Parse a broadcastable message back into a message
     * @param broadcastMessage The broadcastable message to parse
     * @returns The parsed message
     */
    async fromBroadcastFormat(broadcastMessage) {
        const { type, payload, serializedSignatures, signers, metadata } = broadcastMessage;
        if (type !== this.type) {
            throw new Error(`Invalid message type, expected ${this.type}`);
        }
        this.payload = payload;
        this.signatures = (0, iface_1.deserializeSignatures)(serializedSignatures);
        this.signers = signers || [];
        this.metadata = {
            ...metadata,
            encoding: 'utf8',
        };
        return this.build();
    }
}
exports.BaseMessageBuilder = BaseMessageBuilder;
