// @bitgo/sdk-coin-mon/dist/src/mon.js
"use strict";
/***
 * @developerReferences
 * - Developer Hub: https://docs.monad.xyz/introduction/monad-for-developers
 * - Doc: https://docs.monad.xyz/
 *
 * @coinFullName Monad
 * @coinWebsite https://docs.monad.xyz/
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mon = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class Mon extends abstract_eth_1.AbstractEthLikeNewCoins {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
    }
    static createInstance(bitgo, staticsCoin) {
        return new Mon(bitgo, staticsCoin);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    async buildUnsignedSweepTxnTSS(params) {
        return this.buildUnsignedSweepTxnMPCv2(params);
    }
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].monExplorerApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].monExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
}
exports.Mon = Mon;
