// @bitgo/sdk-coin-cronos/dist/src/cronos.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cronos = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const lib_1 = require("./lib");
const constants_1 = require("./lib/constants");
const utils_1 = require("./lib/utils");
/**
 *
 * Full Name: Cronos
 * Website: https://cronos-pos.org/
 * Docs: https://docs.cronos-pos.org/
 * GitHub : https://github.com/crypto-org-chain/chain-main
 */
class Cronos extends abstract_cosmos_1.CosmosCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
        this._utils = new utils_1.Utils();
    }
    static createInstance(bitgo, staticsCoin) {
        return new Cronos(bitgo, staticsCoin);
    }
    /** @inheritDoc **/
    getBaseFactor() {
        return Math.pow(10, this._staticsCoin.decimalPlaces);
    }
    /** @inheritDoc **/
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    /** @inheritDoc **/
    isValidAddress(address) {
        return this._utils.isValidAddress(address) || this._utils.isValidValidatorAddress(address);
    }
    /** @inheritDoc **/
    getDenomination() {
        return statics_1.BaseUnit.CRONOS;
    }
    /** @inheritDoc **/
    getGasAmountDetails() {
        return {
            gasAmount: constants_1.GAS_AMOUNT,
            gasLimit: constants_1.GAS_LIMIT,
        };
    }
    /** @inheritDoc **/
    getKeyPair(publicKey) {
        return new lib_1.KeyPair({ pub: publicKey });
    }
    /** @inheritDoc **/
    getPublicNodeUrl() {
        return sdk_core_1.Environments[this.bitgo.getEnv()].cronosNodeUrl;
    }
    /** @inheritDoc **/
    getAddressFromPublicKey(pubKey) {
        return new lib_1.KeyPair({ pub: pubKey }).getAddress();
    }
}
exports.Cronos = Cronos;
