// @bitgo/public-types/dist/src/schema/transactions/coinSpecific.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinSpecificBuildParams = void 0;
const t = __importStar(require("io-ts"));
const utils_1 = require("../../utils");
const BaseBuildParams = {
    type: t.string,
    nonce: utils_1.optionalNumber,
    memo: t.partial({
        type: utils_1.optionalString,
        value: utils_1.optionalString,
    }),
    comment: utils_1.optionalString,
    ledgerSequenceDelta: utils_1.optionalNumber,
    lastLedgerSequence: utils_1.optionalNumber,
    feeRate: utils_1.optionalStringOrNumber,
    messageKey: utils_1.optionalString,
    previewPendingTxs: utils_1.optionalBoolean,
    pendingApprovalId: utils_1.optionalString,
    sequenceId: utils_1.optionalString,
};
const ConsolidateBuildParams = {
    ...BaseBuildParams,
    consolidateId: utils_1.optionalString,
    senderAddress: utils_1.optionalString,
};
exports.CoinSpecificBuildParams = {
    utxo: {
        locktime: utils_1.optionalNumber,
        limit: utils_1.optionalNumber,
        bulk: utils_1.optionalBoolean,
    },
    avaxp: {
        locktime: utils_1.optionalNumber,
        sourceChain: utils_1.optionalString,
        destinationChain: utils_1.optionalString,
        unspents: utils_1.optionalStringArray,
    },
    dot: {
        ...ConsolidateBuildParams,
        sequenceId: utils_1.optionalString,
        senderAddressIndex: utils_1.optionalNumber,
        nonce: utils_1.optionalNumber,
        proxyType: utils_1.optionalString,
        forceProxyType: utils_1.optionalString,
        owner: utils_1.optionalUnknown,
        delay: utils_1.optionalUnknown,
        index: utils_1.optionalUnknown,
        validityWindow: utils_1.optionalUnknown,
        version: utils_1.optionalUnknown,
        referenceBlock: utils_1.optionalUnknown,
    },
    sol: {
        tokenName: utils_1.optionalString,
    },
};
