// @bitgo/public-types/dist/src/schema/transactions/tronTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TronTransaction = exports.TronWithdrawExpireUnfreezeTransaction = exports.TronClaimRewardsTransaction = exports.TronUndelegateResourceTransaction = exports.TronDelegateResourceTransaction = exports.TronVoteTransaction = exports.TronUnfreezeTransaction = exports.TronFreezeTransaction = exports.TronResourceType = void 0;
const t = __importStar(require("io-ts"));
const utils_1 = require("../../utils");
exports.TronResourceType = t.keyof({
    BANDWIDTH: 1,
    ENERGY: 1,
});
exports.TronFreezeTransaction = t.type({
    type: t.literal("freeze"),
    owner_address: t.string,
    frozen_balance: t.string,
    resource: exports.TronResourceType,
});
exports.TronUnfreezeTransaction = t.type({
    type: t.literal("unfreeze"),
    owner_address: t.string,
    unfreeze_balance: t.string,
    resource: exports.TronResourceType,
});
exports.TronVoteTransaction = t.type({
    type: t.literal("vote"),
    owner_address: t.string,
    votes: t.array(t.type({
        vote_address: t.string,
        vote_count: t.string,
    })),
});
exports.TronDelegateResourceTransaction = t.intersection([
    t.type({
        type: t.literal("delegateResource"),
        owner_address: t.string,
        resource: exports.TronResourceType,
        receiver_address: t.string,
        balance: t.string,
    }),
    t.partial({
        lock: t.boolean,
        lock_period: utils_1.optionalNumber,
    }),
]);
exports.TronUndelegateResourceTransaction = t.type({
    type: t.literal("undelegateResource"),
    owner_address: t.string,
    resource: exports.TronResourceType,
    receiver_address: t.string,
    balance: t.string,
});
exports.TronClaimRewardsTransaction = t.type({
    type: t.literal("claimRewards"),
    owner_address: t.string,
});
exports.TronWithdrawExpireUnfreezeTransaction = t.type({
    type: t.literal("withdrawExpireUnfreeze"),
    owner_address: t.string,
});
exports.TronTransaction = t.union([
    exports.TronFreezeTransaction,
    exports.TronUnfreezeTransaction,
    exports.TronVoteTransaction,
    exports.TronDelegateResourceTransaction,
    exports.TronUndelegateResourceTransaction,
    exports.TronClaimRewardsTransaction,
    exports.TronWithdrawExpireUnfreezeTransaction,
]);
