// @bitgo/sdk-core/dist/test/unit/account-lib/baseCoin/messages/baseMessage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const statics_1 = require("@bitgo/statics");
const sinon_1 = __importDefault(require("sinon"));
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../../../src");
const fixtures_1 = require("./fixtures");
describe('Base Message', () => {
    let coinConfig;
    beforeEach(() => {
        coinConfig = sinon_1.default.createStubInstance(statics_1.BaseCoin);
    });
    it('should initialize with default values', () => {
        const message = new fixtures_1.TestMessage({
            coinConfig,
            payload: '',
        });
        should_1.default.equal(message.getType(), src_1.MessageStandardType.UNKNOWN);
        should_1.default.equal(message.getPayload(), '');
        should_1.default.deepEqual(message.getMetadata(), {});
        should_1.default.deepEqual(message.getSignatures(), []);
        should_1.default.deepEqual(message.getSigners(), []);
    });
    it('should initialize with provided values', () => {
        const { payload, type, metadata, signers, signatures } = fixtures_1.messageSamples.eip191;
        const message = new fixtures_1.TestMessage({
            coinConfig,
            payload,
            type,
            metadata,
            signers,
            signatures,
        });
        should_1.default.equal(message.getType(), type);
        should_1.default.equal(message.getPayload(), payload);
        should_1.default.deepEqual(message.getMetadata(), metadata);
        should_1.default.deepEqual(message.getSignatures(), signatures);
        should_1.default.deepEqual(message.getSigners(), signers);
    });
    describe('Getters and Setters', () => {
        let message;
        beforeEach(() => {
            message = new fixtures_1.TestMessage({
                coinConfig,
                payload: 'test',
            });
        });
        it('should handle adding and getting signers', () => {
            const signer1 = '0xabc123';
            const signer2 = '0xdef456';
            message.addSigner(signer1);
            should_1.default.deepEqual(message.getSigners(), [signer1]);
            message.addSigner(signer2);
            should_1.default.deepEqual(message.getSigners(), [signer1, signer2]);
            // Adding a duplicate signer should not add it again
            message.addSigner(signer1);
            should_1.default.deepEqual(message.getSigners(), [signer1, signer2]);
            // Set signers should replace all existing signers
            const newSigners = ['0x111', '0x222'];
            message.setSigners(newSigners);
            should_1.default.deepEqual(message.getSigners(), newSigners);
        });
        it('should handle adding and getting signatures', () => {
            const sig1 = {
                publicKey: { pub: 'pub1' },
                signature: Buffer.from('signature1'),
            };
            const sig2 = {
                publicKey: { pub: 'pub2' },
                signature: Buffer.from('signature2'),
            };
            message.addSignature(sig1);
            should_1.default.deepEqual(message.getSignatures(), [sig1]);
            message.addSignature(sig2);
            should_1.default.deepEqual(message.getSignatures(), [sig1, sig2]);
            // Set signatures should replace all existing signatures
            const newSignatures = [
                {
                    publicKey: { pub: 'pub3' },
                    signature: Buffer.from('sig3'),
                },
                {
                    publicKey: { pub: 'pub4' },
                    signature: Buffer.from('sig4'),
                },
            ];
            message.setSignatures(newSignatures);
            should_1.default.deepEqual(message.getSignatures(), newSignatures);
        });
        it('should return copies of arrays to prevent mutation', () => {
            const signers = ['addr1', 'addr2'];
            const signatures = [
                {
                    publicKey: { pub: 'pub1' },
                    signature: Buffer.from('sig1'),
                },
                {
                    publicKey: { pub: 'pub1' },
                    signature: Buffer.from('sig1'),
                },
            ];
            message.setSigners(signers);
            message.setSignatures(signatures);
            // Modifying the returned arrays should not affect the internal state
            const returnedSigners = message.getSigners();
            const returnedSignatures = message.getSignatures();
            returnedSigners.push('addr3');
            returnedSignatures.push({
                publicKey: { pub: 'pub3' },
                signature: Buffer.from('sig3'),
            });
            should_1.default.deepEqual(message.getSigners(), signers);
            should_1.default.deepEqual(message.getSignatures(), signatures);
        });
    });
    describe('getSignablePayload', () => {
        it('should return the payload as buffer if signablePayload is not set', async () => {
            const payload = 'test payload';
            const message = new fixtures_1.TestMessage({
                coinConfig,
                payload,
            });
            const result = await message.getSignablePayload();
            should_1.default.deepEqual(result, Buffer.from(payload));
        });
    });
    describe('toBroadcastFormat', () => {
        it('should create a proper broadcastable format with all fields', async () => {
            const { payload, type, metadata, signers, signatures } = fixtures_1.messageSamples.eip191;
            const message = new fixtures_1.TestMessage({
                coinConfig,
                payload,
                type,
                metadata,
                signers,
                signatures,
            });
            const broadcastFormat = await message.toBroadcastFormat();
            should_1.default.deepEqual(broadcastFormat, {
                type,
                payload,
                serializedSignatures: (0, src_1.serializeSignatures)(signatures),
                signers,
                metadata,
                signablePayload: 'SGVsbG8gQml0R28h',
            });
        });
        it('should perform a deep copy of metadata to prevent mutation', async () => {
            const nestedMetadata = {
                version: '1.0',
                settings: {
                    chainId: 1,
                    gasLimit: 21000,
                },
            };
            const message = new fixtures_1.TestMessage({
                coinConfig,
                payload: 'test',
                metadata: nestedMetadata,
                signers: ['addr1'],
                signatures: [
                    {
                        publicKey: { pub: 'pub1' },
                        signature: Buffer.from('sig1'),
                    },
                ],
            });
            const broadcastFormat = await message.toBroadcastFormat();
            // The metadata in the broadcast format should be a deep copy
            should_1.default.deepEqual(broadcastFormat.metadata, nestedMetadata);
            // But it should not be the same object reference
            should_1.default.notEqual(broadcastFormat.metadata, nestedMetadata);
            // Modifying the original should not affect the broadcasted version
            nestedMetadata.settings.gasLimit = 50000;
            should_1.default.notEqual(broadcastFormat.metadata.settings.gasLimit, 50000);
        });
    });
    describe('toBroadcastString', () => {
        it('should serialize the broadcastable format to hex string', async () => {
            const { payload, type, metadata, signers, signatures } = fixtures_1.messageSamples.eip191;
            const message = new fixtures_1.TestMessage({
                coinConfig,
                payload,
                type,
                metadata,
                signers,
                signatures,
            });
            const expectedBroadcastString = {
                type,
                payload,
                serializedSignatures: (0, src_1.serializeSignatures)(signatures),
                signers,
                metadata,
                signablePayload: 'SGVsbG8gQml0R28h',
            };
            const broadcastHex = await message.toBroadcastString();
            const broadcastString = Buffer.from(broadcastHex, 'hex').toString();
            const parsed = JSON.parse(broadcastString);
            should_1.default.deepEqual(parsed, expectedBroadcastString);
        });
    });
});
