// @bitgo/sdk-core/dist/test/unit/account-lib/baseCoin/messages/baseMessageBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const statics_1 = require("@bitgo/statics");
const sinon_1 = __importDefault(require("sinon"));
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../../../src");
const fixtures_1 = require("./fixtures");
describe('Base Message Builder Factory', () => {
    let factory;
    let mockCoinConfig;
    beforeEach(() => {
        mockCoinConfig = sinon_1.default.createStubInstance(statics_1.BaseCoin);
        factory = new fixtures_1.TestMessageBuilderFactory(mockCoinConfig);
    });
    describe('getMessageBuilder', () => {
        it('should create a message builder for EIP191 type', () => {
            const builder = factory.getMessageBuilder(src_1.MessageStandardType.EIP191);
            should_1.default.exist(builder);
            should_1.default.equal(builder.getType(), src_1.MessageStandardType.EIP191);
        });
        it('should create a message builder for UNKNOWN type', () => {
            const builder = factory.getMessageBuilder(src_1.MessageStandardType.UNKNOWN);
            should_1.default.exist(builder);
            should_1.default.equal(builder.getType(), src_1.MessageStandardType.UNKNOWN);
        });
    });
    describe('fromBroadcastFormat', () => {
        it('should create a builder from a broadcast message', () => {
            const broadcastMessage = {
                type: src_1.MessageStandardType.EIP191,
                payload: 'hello world',
                serializedSignatures: [
                    {
                        publicKey: 'pub1',
                        signature: 'sig1',
                    },
                ],
                signers: ['addr1'],
                metadata: { version: '1.0' },
            };
            const builder = factory.fromBroadcastFormat(broadcastMessage);
            should_1.default.exist(builder);
            // Since the TestMessageBuilder always returns the same type that was passed to constructor
            should_1.default.equal(builder.getType(), src_1.MessageStandardType.EIP191);
        });
    });
    describe('fromBroadcastString', () => {
        it('should parse a broadcast message string and create the appropriate builder', () => {
            const broadcastMessage = {
                type: src_1.MessageStandardType.EIP191,
                payload: 'test message',
                serializedSignatures: [
                    {
                        publicKey: 'pub1',
                        signature: 'sig1',
                    },
                    {
                        publicKey: 'pub2',
                        signature: 'sig2',
                    },
                ],
                signers: ['addr1', 'addr2'],
                metadata: { chainId: 1 },
            };
            const broadcastString = JSON.stringify(broadcastMessage);
            const broadcastHex = Buffer.from(broadcastString).toString('hex');
            const builder = factory.fromBroadcastString(broadcastHex);
            should_1.default.exist(builder);
            // Since the TestMessageBuilder always returns the same type that was passed to constructor
            should_1.default.equal(builder.getType(), src_1.MessageStandardType.EIP191);
        });
        it('should handle broadcast messages with different types', () => {
            const broadcastMessage = {
                type: src_1.MessageStandardType.UNKNOWN,
                payload: 'unknown message',
                serializedSignatures: [
                    {
                        publicKey: 'pub1',
                        signature: 'sig1',
                    },
                ],
                signers: ['addr1'],
            };
            const broadcastString = JSON.stringify(broadcastMessage);
            const broadcastHex = Buffer.from(broadcastString).toString('hex');
            const builder = factory.fromBroadcastString(broadcastHex);
            should_1.default.exist(builder);
            should_1.default.equal(builder.getType(), src_1.MessageStandardType.UNKNOWN);
        });
        it('should handle broadcast messages without optional fields', () => {
            const broadcastMessage = {
                type: src_1.MessageStandardType.EIP191,
                payload: 'minimal message',
                signatures: [],
                signers: [],
            };
            const broadcastString = JSON.stringify(broadcastMessage);
            const broadcastHex = Buffer.from(broadcastString).toString('hex');
            const builder = factory.fromBroadcastString(broadcastHex);
            should_1.default.exist(builder);
            should_1.default.equal(builder.getType(), src_1.MessageStandardType.EIP191);
        });
    });
});
