// @bitgo/sdk-core/dist/test/unit/account-lib/baseCoin/messages/baseMessageBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const statics_1 = require("@bitgo/statics");
const sinon_1 = __importDefault(require("sinon"));
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../../../src");
const fixtures_1 = require("./fixtures");
describe('Base Message Builder', () => {
    let builder;
    let mockCoinConfig;
    beforeEach(() => {
        mockCoinConfig = sinon_1.default.createStubInstance(statics_1.BaseCoin);
        builder = new fixtures_1.TestMessageBuilder(mockCoinConfig, src_1.MessageStandardType.EIP191);
    });
    it('should initialize with default values', () => {
        should_1.default.equal(builder.getPayload(), '');
        should_1.default.equal(builder.getType(), src_1.MessageStandardType.EIP191);
        should_1.default.deepEqual(builder.getSignatures(), []);
        should_1.default.deepEqual(builder.getSigners(), []);
        should_1.default.deepEqual(builder.getMetadata(), {});
        should_1.default.equal(builder.getDigest(), undefined);
    });
    it('should set and get payload', () => {
        const payload = 'test payload';
        builder.setPayload(payload);
        should_1.default.equal(builder.getPayload(), payload);
    });
    it('should set and get metadata', () => {
        const metadata = { key: 'value', num: 123 };
        builder.setMetadata(metadata);
        should_1.default.deepEqual(builder.getMetadata(), metadata);
    });
    it('should set and get type', () => {
        const type = src_1.MessageStandardType.UNKNOWN;
        builder.setType(type);
        should_1.default.equal(builder.getType(), type);
    });
    it('should set and get signatures', () => {
        const signatures = [
            {
                publicKey: { pub: 'pubKey1' },
                signature: Buffer.from('sig1'),
            },
            {
                publicKey: { pub: 'pubKey2' },
                signature: Buffer.from('sig2'),
            },
            {
                publicKey: { pub: 'pubKey3' },
                signature: Buffer.from('sig3'),
            },
        ];
        builder.setSignatures(signatures);
        should_1.default.deepEqual(builder.getSignatures(), signatures);
    });
    it('should set and get signers', () => {
        const signers = ['address1', 'address2', 'address3'];
        builder.setSigners(signers);
        should_1.default.deepEqual(builder.getSigners(), signers);
    });
    it('should set and get digest', () => {
        const digest = '0x1234abcd';
        builder.setDigest(digest);
        should_1.default.equal(builder.getDigest(), digest);
    });
    it('should build a message with the correct properties', async () => {
        const payload = 'test message';
        const metadata = { foo: 'bar', encoding: 'utf8' };
        const signatures = [
            {
                publicKey: { pub: 'pubKey1' },
                signature: Buffer.from('sig1'),
            },
            {
                publicKey: { pub: 'pubKey2' },
                signature: Buffer.from('sig2'),
            },
        ];
        const signers = ['addr1', 'addr2'];
        builder
            .setType(src_1.MessageStandardType.EIP191)
            .setPayload(payload)
            .setMetadata(metadata)
            .setSignatures(signatures)
            .setSigners(signers);
        const message = await builder.build();
        should_1.default.equal(message.getType(), src_1.MessageStandardType.EIP191);
        should_1.default.equal(message.getPayload(), payload);
        should_1.default.deepEqual(message.getMetadata(), metadata);
        should_1.default.deepEqual(message.getSignatures(), signatures);
        should_1.default.deepEqual(message.getSigners(), signers);
        const signablePayload = await message.getSignablePayload();
        should_1.default.deepEqual(signablePayload, Buffer.from(payload));
    });
    it('should correctly handle toBroadcastFormat', async () => {
        const payload = 'hello world';
        const metadata = { version: '1.0', encoding: 'utf8' };
        const signatures = [
            {
                publicKey: { pub: 'pubKey1' },
                signature: Buffer.from('sig1'),
            },
        ];
        const signers = ['addr1'];
        builder
            .setType(src_1.MessageStandardType.EIP191)
            .setPayload(payload)
            .setMetadata(metadata)
            .setSignatures(signatures)
            .setSigners(signers);
        const message = await builder.build();
        const broadcastFormat = await message.toBroadcastFormat();
        const expectedBroadcastFormat = {
            type: src_1.MessageStandardType.EIP191,
            payload: payload,
            serializedSignatures: (0, src_1.serializeSignatures)(signatures),
            signers: signers,
            metadata: metadata,
            signablePayload: 'aGVsbG8gd29ybGQ=',
        };
        should_1.default.deepEqual(broadcastFormat, expectedBroadcastFormat);
    });
    it('should correctly handle fromBroadcastFormat', async () => {
        const broadcastMessage = {
            type: src_1.MessageStandardType.EIP191,
            payload: 'broadcast test',
            signatures: [
                {
                    publicKey: { pub: 'pubKey1' },
                    signature: Buffer.from('sig1'),
                },
                {
                    publicKey: { pub: 'pubKey2' },
                    signature: Buffer.from('sig2'),
                },
            ],
            signers: ['addr1', 'addr2'],
            metadata: { chainId: 1, encoding: 'utf8' },
        };
        const message = await builder.fromBroadcastFormat(broadcastMessage);
        should_1.default.equal(message.getType(), broadcastMessage.type);
        should_1.default.equal(message.getPayload(), broadcastMessage.payload);
        should_1.default.deepEqual(message.getSignatures(), broadcastMessage.signatures);
        should_1.default.deepEqual(message.getSigners(), broadcastMessage.signers);
        should_1.default.deepEqual(message.getMetadata(), broadcastMessage.metadata);
    });
    it('should correctly handle toBroadcastString', async () => {
        const payload = 'serialize me';
        const signatures = [
            {
                publicKey: { pub: 'pubKey1' },
                signature: Buffer.from('sig1'),
            },
        ];
        const signers = ['addr1'];
        builder.setType(src_1.MessageStandardType.EIP191).setPayload(payload).setSignatures(signatures).setSigners(signers);
        const message = await builder.build();
        const broadcastHex = await message.toBroadcastString();
        const expectedJson = JSON.stringify({
            type: src_1.MessageStandardType.EIP191,
            payload: payload,
            serializedSignatures: (0, src_1.serializeSignatures)(signatures),
            signers: signers,
            metadata: { encoding: 'utf8' },
            signablePayload: 'c2VyaWFsaXplIG1l',
        });
        const expectedHex = Buffer.from(expectedJson).toString('hex');
        should_1.default.equal(broadcastHex, expectedHex);
    });
});
