// @bitgo/sdk-core/dist/test/unit/account-lib/baseCoin/messages/fixtures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageSamples = exports.TestMessage = exports.TestMessageBuilder = exports.TestMessageBuilderFactory = void 0;
const src_1 = require("../../../../../src");
class TestMessageBuilderFactory extends src_1.BaseMessageBuilderFactory {
    constructor(coinConfig) {
        super(coinConfig);
    }
    getMessageBuilder(type) {
        return new TestMessageBuilder(this.coinConfig, type);
    }
}
exports.TestMessageBuilderFactory = TestMessageBuilderFactory;
class TestMessageBuilder extends src_1.BaseMessageBuilder {
    constructor(coinConfig, type = src_1.MessageStandardType.UNKNOWN) {
        super(coinConfig, type);
    }
    async buildMessage(options) {
        return new TestMessage(options);
    }
    async fromBroadcastFormat(broadcastMessage) {
        this.setType(broadcastMessage.type);
        this.setPayload(broadcastMessage.payload);
        this.setSignatures(broadcastMessage.signatures || []);
        this.setSigners(broadcastMessage.signers || []);
        if (broadcastMessage.metadata) {
            this.setMetadata(broadcastMessage.metadata);
        }
        return this.build();
    }
}
exports.TestMessageBuilder = TestMessageBuilder;
class TestMessage extends src_1.BaseMessage {
    constructor(options) {
        super(options);
    }
    async getSignablePayload() {
        this.signablePayload = Buffer.from(this.payload);
        return this.signablePayload;
    }
}
exports.TestMessage = TestMessage;
exports.messageSamples = {
    eip191: {
        payload: 'Hello BitGo!',
        type: src_1.MessageStandardType.EIP191,
        metadata: { chainId: 1 },
        signers: ['0x1234567890abcdef1234567890abcdef12345678'],
        signatures: [
            {
                publicKey: { pub: '0x1234567890abcdef1234567890abcdef12345678' },
                signature: Buffer.from('abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890', 'hex'),
            },
        ],
    },
    unknown: {
        payload: 'Unknown message type',
        type: src_1.MessageStandardType.UNKNOWN,
        metadata: { version: '1.0' },
        signers: ['12345'],
        signatures: [
            {
                publicKey: { pub: '12345' },
                signature: Buffer.from('67890'),
            },
        ],
    },
};
