// @bitgo/sdk-coin-hash/dist/src/hashToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HashToken = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const hash_1 = require("./hash");
const utils_1 = require("./lib/utils");
const lib_1 = require("./lib");
class HashToken extends hash_1.Hash {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = statics_1.coins.get(tokenConfig.coin);
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
        this._utils = new utils_1.HashUtils(staticsCoin.network.type);
    }
    static createTokenConstructor(config) {
        return (bitgo) => new HashToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.cosmos.tokens, ...statics_1.tokens.testnet.cosmos.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            if (token.coin === 'hash' || token.coin === 'thash') {
                const tokenConstructor = HashToken.createTokenConstructor(token);
                tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
            }
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get denom() {
        return this.tokenConfig.denom;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Hash Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    getAddressFromPublicKey(publicKey) {
        const networkType = this.network === 'Mainnet' ? sdk_core_1.AddressFormat.mainnet : sdk_core_1.AddressFormat.testnet;
        return new lib_1.KeyPair({ pub: publicKey }).getAddress(networkType);
    }
}
exports.HashToken = HashToken;
