// @bitgo/sdk-coin-hash/dist/src/hash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hash = void 0;
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
const utils_1 = require("./lib/utils");
class Hash extends abstract_cosmos_1.CosmosCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
        this._utils = new utils_1.HashUtils(statics_1.NetworkType.MAINNET);
    }
    static createInstance(bitgo, staticsCoin) {
        return new Hash(bitgo, staticsCoin);
    }
    /** @inheritDoc **/
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    /** @inheritDoc **/
    getBaseFactor() {
        return 1e9;
    }
    /** @inheritDoc **/
    isValidAddress(address) {
        return this._utils.isValidAddress(address) || this._utils.isValidValidatorAddress(address);
    }
    /** @inheritDoc **/
    getPublicNodeUrl() {
        return sdk_core_1.Environments[this.bitgo.getEnv()].hashNodeUrl;
    }
    /** @inheritDoc **/
    getDenomination() {
        return statics_1.BaseUnit.HASH;
    }
    /** @inheritDoc **/
    getGasAmountDetails() {
        return {
            gasAmount: '5000000000',
            gasLimit: 250000,
        };
    }
    /** @inheritDoc **/
    getKeyPair(publicKey) {
        return new lib_1.KeyPair({ pub: publicKey });
    }
    /** @inheritDoc **/
    getAddressFromPublicKey(publicKey) {
        return new lib_1.KeyPair({ pub: publicKey }).getAddress(sdk_core_1.AddressFormat.mainnet);
    }
    /** @inheritDoc **/
    valuelessTransferAllowed() {
        return true;
    }
}
exports.Hash = Hash;
