// @bitgo/sdk-coin-tempo/dist/src/lib/tip20Abi.js
"use strict";
/**
 * TIP-20 Token Standard ABI
 *
 * TIP-20 is Tempo's token standard, similar to ERC-20 but with:
 * - 6 decimal places (instead of 18)
 * - transferWithMemo function for attaching metadata
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TIP20_ABI = exports.TIP20_TRANSFER_WITH_MEMO_ABI = void 0;
/**
 * TIP-20 transferWithMemo ABI
 * Standard function for TIP-20 token transfers with memo field
 */
exports.TIP20_TRANSFER_WITH_MEMO_ABI = [
    {
        type: 'function',
        name: 'transferWithMemo',
        inputs: [
            { name: 'to', type: 'address' },
            { name: 'amount', type: 'uint256' },
            { name: 'memo', type: 'bytes32' },
        ],
        outputs: [{ name: '', type: 'bool' }],
        stateMutability: 'nonpayable',
    },
];
/**
 * Standard TIP-20 token ABI (similar to ERC-20)
 */
exports.TIP20_ABI = [
    {
        type: 'function',
        name: 'balanceOf',
        inputs: [{ name: 'account', type: 'address' }],
        outputs: [{ name: '', type: 'uint256' }],
        stateMutability: 'view',
    },
    {
        type: 'function',
        name: 'transfer',
        inputs: [
            { name: 'to', type: 'address' },
            { name: 'amount', type: 'uint256' },
        ],
        outputs: [{ name: '', type: 'bool' }],
        stateMutability: 'nonpayable',
    },
    {
        type: 'function',
        name: 'transferFrom',
        inputs: [
            { name: 'from', type: 'address' },
            { name: 'to', type: 'address' },
            { name: 'amount', type: 'uint256' },
        ],
        outputs: [{ name: '', type: 'bool' }],
        stateMutability: 'nonpayable',
    },
    ...exports.TIP20_TRANSFER_WITH_MEMO_ABI,
];
