// @bitgo/sdk-coin-tempo/dist/src/lib/transactionBuilder.js
"use strict";
/**
 * TIP-20 Transaction Builder
 *
 * Unified builder for TIP-20 transactions supporting:
 * - Single or batch operations
 * - Per-operation memos for tracking
 * - Custom fee token selection
 * - EIP-7702 Account Abstraction (type 0x76)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tip20TransactionBuilder = void 0;
const abstract_eth_1 = require("@bitgo/abstract-eth");
const sdk_core_1 = require("@bitgo/sdk-core");
const ethers_1 = require("ethers");
const transaction_1 = require("./transaction");
const utils_1 = require("./utils");
const tip20Abi_1 = require("./tip20Abi");
const constants_1 = require("./constants");
/**
 * Transaction Builder for TIP-20 tokens on Tempo blockchain
 * Extends abstract-eth TransactionBuilder with Tempo-specific features
 */
class Tip20TransactionBuilder extends abstract_eth_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.operations = [];
    }
    /**
     * Implement the transfer method from abstract class
     * Note: For TIP-20 transactions, use addOperation() instead
     */
    transfer(data, isFirstSigner) {
        const transferBuilder = new abstract_eth_1.TransferBuilder(undefined, isFirstSigner);
        if (data) {
            transferBuilder.data(data);
        }
        this._transfer = transferBuilder;
        return transferBuilder;
    }
    /**
     * Validate the transaction has all required fields for Tempo AA transactions.
     * Overrides parent class validation since AA transactions use a different model
     * (operations-based rather than single contract address).
     *
     * @throws BuildTransactionError if validation fails
     */
    validateTransaction() {
        if (this.operations.length === 0) {
            throw new sdk_core_1.BuildTransactionError('At least one operation is required to build a transaction');
        }
        if (this._nonce === undefined) {
            throw new sdk_core_1.BuildTransactionError('Nonce is required to build a transaction');
        }
        if (this._gas === undefined) {
            throw new sdk_core_1.BuildTransactionError('Gas limit is required to build a transaction');
        }
        if (this._maxFeePerGas === undefined) {
            throw new sdk_core_1.BuildTransactionError('maxFeePerGas is required to build a transaction');
        }
        if (this._maxPriorityFeePerGas === undefined) {
            throw new sdk_core_1.BuildTransactionError('maxPriorityFeePerGas is required to build a transaction');
        }
    }
    /** @inheritdoc */
    validateRawTransaction(rawTransaction) {
        if (typeof rawTransaction === 'string' && (0, utils_1.isTip20Transaction)(rawTransaction)) {
            try {
                ethers_1.ethers.utils.RLP.decode('0x' + rawTransaction.slice(4));
                return;
            }
            catch (e) {
                throw new sdk_core_1.ParseTransactionError(`Failed to RLP decode TIP-20 transaction: ${e}`);
            }
        }
        super.validateRawTransaction(rawTransaction);
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction, isFirstSigner) {
        if (!rawTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Raw transaction is empty');
        }
        if ((0, utils_1.isTip20Transaction)(rawTransaction)) {
            return this.fromTip20Transaction(rawTransaction);
        }
        return super.fromImplementation(rawTransaction, isFirstSigner);
    }
    /**
     * Deserialize a type 0x76 transaction and restore builder state.
     * RLP field layout mirrors buildBaseRlpData() in transaction.ts.
     */
    fromTip20Transaction(rawTransaction) {
        try {
            const rlpHex = '0x' + rawTransaction.slice(4);
            const decoded = ethers_1.ethers.utils.RLP.decode(rlpHex);
            if (!Array.isArray(decoded) || decoded.length < 13) {
                throw new sdk_core_1.ParseTransactionError('Invalid TIP-20 transaction: unexpected RLP structure');
            }
            const parseBigInt = (hex) => (!hex || hex === '0x' ? 0n : BigInt(hex));
            const parseHexInt = (hex) => (!hex || hex === '0x' ? 0 : parseInt(hex, 16));
            const chainId = parseHexInt(decoded[0]);
            const maxPriorityFeePerGas = parseBigInt(decoded[1]);
            const maxFeePerGas = parseBigInt(decoded[2]);
            const gas = parseBigInt(decoded[3]);
            const callsTuples = decoded[4];
            const nonce = parseHexInt(decoded[7]);
            const feeTokenRaw = decoded[10];
            const calls = callsTuples.map((tuple) => ({
                to: tuple[0],
                value: parseBigInt(tuple[1]),
                data: tuple[2],
            }));
            const operations = calls.map((call) => this.decodeCallToOperation(call));
            let signature;
            if (decoded.length >= 14 && decoded[13] && decoded[13].length > 2) {
                const sigBytes = ethers_1.ethers.utils.arrayify(decoded[13]);
                if (sigBytes.length === 65) {
                    const r = ethers_1.ethers.utils.hexlify(sigBytes.slice(0, 32));
                    const s = ethers_1.ethers.utils.hexlify(sigBytes.slice(32, 64));
                    const v = sigBytes[64];
                    const yParity = v > 1 ? v - 27 : v;
                    signature = { r, s, yParity };
                }
            }
            const feeToken = feeTokenRaw && feeTokenRaw !== '0x' ? feeTokenRaw : undefined;
            const txRequest = {
                type: constants_1.AA_TRANSACTION_TYPE,
                chainId,
                nonce,
                maxFeePerGas,
                maxPriorityFeePerGas,
                gas,
                calls,
                accessList: [],
                feeToken,
            };
            this._nonce = nonce;
            this._gas = gas;
            this._maxFeePerGas = maxFeePerGas;
            this._maxPriorityFeePerGas = maxPriorityFeePerGas;
            this._feeToken = feeToken;
            this.operations = operations;
            this._restoredSignature = signature;
            const tx = new transaction_1.Tip20Transaction(this._coinConfig, txRequest, operations);
            if (signature) {
                tx.setSignature(signature);
            }
            return tx;
        }
        catch (e) {
            if (e instanceof sdk_core_1.ParseTransactionError)
                throw e;
            throw new sdk_core_1.ParseTransactionError(`Failed to deserialize TIP-20 transaction: ${e}`);
        }
    }
    /**
     * Decode a single AA call's data back into a Tip20Operation.
     * Expects the call data to encode transferWithMemo(address, uint256, bytes32).
     */
    decodeCallToOperation(call) {
        const iface = new ethers_1.ethers.utils.Interface(tip20Abi_1.TIP20_TRANSFER_WITH_MEMO_ABI);
        try {
            const decoded = iface.decodeFunctionData('transferWithMemo', call.data);
            const toAddress = decoded[0];
            const amountUnits = BigInt(decoded[1].toString());
            const memoBytes32 = decoded[2];
            const amount = (0, utils_1.tip20UnitsToAmount)(amountUnits);
            const stripped = ethers_1.ethers.utils.stripZeros(memoBytes32);
            const memo = stripped.length > 0 ? ethers_1.ethers.utils.toUtf8String(stripped) : undefined;
            return { token: call.to, to: toAddress, amount, memo };
        }
        catch {
            return { token: call.to, to: call.to, amount: (0, utils_1.tip20UnitsToAmount)(call.value) };
        }
    }
    /**
     * Build the transaction from configured TIP-20 operations and transaction parameters.
     * Signs with _sourceKeyPair if it has been set via sign({ key }).
     */
    async buildImplementation() {
        if (this._nonce === undefined ||
            this._gas === undefined ||
            this._maxFeePerGas === undefined ||
            this._maxPriorityFeePerGas === undefined) {
            throw new sdk_core_1.BuildTransactionError('Transaction validation failed: missing required fields');
        }
        this.fee({
            fee: this._maxFeePerGas.toString(),
            gasLimit: this._gas.toString(),
            eip1559: {
                maxFeePerGas: this._maxFeePerGas.toString(),
                maxPriorityFeePerGas: this._maxPriorityFeePerGas.toString(),
            },
        });
        const calls = this.operations.map((op) => this.operationToCall(op));
        const txRequest = {
            type: constants_1.AA_TRANSACTION_TYPE,
            chainId: this._common.chainIdBN().toNumber(),
            nonce: this._nonce,
            maxFeePerGas: this._maxFeePerGas,
            maxPriorityFeePerGas: this._maxPriorityFeePerGas,
            gas: this._gas,
            calls,
            accessList: [],
            feeToken: this._feeToken,
        };
        const tx = new transaction_1.Tip20Transaction(this._coinConfig, txRequest, this.operations);
        if (this._sourceKeyPair && this._sourceKeyPair.getKeys().prv) {
            const prv = this._sourceKeyPair.getKeys().prv;
            const unsignedHex = await tx.serialize();
            const msgHash = ethers_1.ethers.utils.keccak256(ethers_1.ethers.utils.arrayify(unsignedHex));
            const signingKey = new ethers_1.ethers.utils.SigningKey('0x' + prv);
            const sig = signingKey.signDigest(ethers_1.ethers.utils.arrayify(msgHash));
            tx.setSignature({
                r: sig.r,
                s: sig.s,
                yParity: sig.recoveryParam ?? 0,
            });
        }
        else if (this._restoredSignature) {
            tx.setSignature(this._restoredSignature);
        }
        return tx;
    }
    /**
     * Add a single operation to the transaction
     * Can be called multiple times to create batch transactions
     *
     * @param operation - TIP-20 operation with token, recipient, amount, and optional memo
     * @returns this builder instance for chaining
     */
    addOperation(operation) {
        this.validateOperation(operation);
        this.operations.push(operation);
        return this;
    }
    /**
     * Set which TIP-20 token will be used to pay transaction fees
     * This is a global setting for the entire transaction
     *
     * @param tokenAddress - Address of the TIP-20 token to use for fees
     * @returns this builder instance for chaining
     */
    feeToken(tokenAddress) {
        if (!(0, utils_1.isValidAddress)(tokenAddress)) {
            throw new sdk_core_1.BuildTransactionError(`Invalid fee token address: ${tokenAddress}`);
        }
        this._feeToken = tokenAddress;
        return this;
    }
    /**
     * Set the transaction nonce
     *
     * @param nonce - Transaction nonce
     * @returns this builder instance for chaining
     */
    nonce(nonce) {
        if (nonce < 0) {
            throw new sdk_core_1.BuildTransactionError(`Invalid nonce: ${nonce}`);
        }
        this._nonce = nonce;
        return this;
    }
    /**
     * Set the gas limit for the transaction
     *
     * @param gas - Gas limit
     * @returns this builder instance for chaining
     */
    gas(gas) {
        const gasValue = typeof gas === 'string' ? BigInt(gas) : gas;
        if (gasValue <= 0n) {
            throw new sdk_core_1.BuildTransactionError(`Invalid gas limit: ${gas}`);
        }
        this._gas = gasValue;
        return this;
    }
    /**
     * Set the maximum fee per gas (EIP-1559)
     *
     * @param maxFeePerGas - Maximum fee per gas in wei
     * @returns this builder instance for chaining
     */
    maxFeePerGas(maxFeePerGas) {
        const feeValue = typeof maxFeePerGas === 'string' ? BigInt(maxFeePerGas) : maxFeePerGas;
        if (feeValue < 0n) {
            throw new sdk_core_1.BuildTransactionError(`Invalid maxFeePerGas: ${maxFeePerGas}`);
        }
        this._maxFeePerGas = feeValue;
        return this;
    }
    /**
     * Set the maximum priority fee per gas (EIP-1559)
     *
     * @param maxPriorityFeePerGas - Maximum priority fee per gas in wei
     * @returns this builder instance for chaining
     */
    maxPriorityFeePerGas(maxPriorityFeePerGas) {
        const feeValue = typeof maxPriorityFeePerGas === 'string' ? BigInt(maxPriorityFeePerGas) : maxPriorityFeePerGas;
        if (feeValue < 0n) {
            throw new sdk_core_1.BuildTransactionError(`Invalid maxPriorityFeePerGas: ${maxPriorityFeePerGas}`);
        }
        this._maxPriorityFeePerGas = feeValue;
        return this;
    }
    /**
     * Get all operations in this transaction
     * @returns Array of TIP-20 operations
     */
    getOperations() {
        return [...this.operations];
    }
    /**
     * Get the fee token address if set
     * @returns Fee token address or undefined
     */
    getFeeToken() {
        return this._feeToken;
    }
    /**
     * Validate a single operation
     * @param operation - Operation to validate
     * @throws BuildTransactionError if invalid
     */
    validateOperation(operation) {
        if (!(0, utils_1.isValidAddress)(operation.token)) {
            throw new sdk_core_1.BuildTransactionError(`Invalid token address: ${operation.token}`);
        }
        if (!(0, utils_1.isValidAddress)(operation.to)) {
            throw new sdk_core_1.BuildTransactionError(`Invalid recipient address: ${operation.to}`);
        }
        if (!(0, utils_1.isValidTip20Amount)(operation.amount)) {
            throw new sdk_core_1.BuildTransactionError(`Invalid amount: ${operation.amount}`);
        }
        if (operation.memo !== undefined && !(0, utils_1.isValidMemoId)(operation.memo)) {
            throw new sdk_core_1.BuildTransactionError(`Invalid memo: must be a non-negative integer`);
        }
    }
    /**
     * Convert a TIP-20 operation to an AA call
     */
    operationToCall(op) {
        const amountInUnits = (0, utils_1.amountToTip20Units)(op.amount);
        const data = (0, utils_1.encodeTip20TransferWithMemo)(op.to, amountInUnits, op.memo);
        return {
            to: op.token,
            data,
            value: 0n,
        };
    }
}
exports.Tip20TransactionBuilder = Tip20TransactionBuilder;
