// @bitgo/sdk-coin-tempo/dist/src/lib/transaction.js
"use strict";
/**
 * TIP-20 Transaction
 *
 * Represents a Tempo Account Abstraction (AA) transaction (type 0x76)
 * Supports single or batch TIP-20 token transfers with memos
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tip20Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const ethers_1 = require("ethers");
const utils_1 = require("./utils");
class Tip20Transaction extends sdk_core_1.BaseTransaction {
    constructor(_coinConfig, request, operations = []) {
        super(_coinConfig);
        this.txRequest = request;
        this._operations = operations;
        this._outputs = operations.map((op) => ({
            address: op.to,
            value: (0, utils_1.amountToTip20Units)(op.amount).toString(),
            coin: op.token,
        }));
        const totalUnits = operations.reduce((sum, op) => sum + (0, utils_1.amountToTip20Units)(op.amount), 0n);
        this._inputs = [{ address: '', value: totalUnits.toString(), coin: _coinConfig.name }];
    }
    get type() {
        return sdk_core_1.TransactionType.Send;
    }
    canSign() {
        return true;
    }
    async serialize(signature) {
        const sig = signature || this._signature;
        return this.serializeTransaction(sig);
    }
    /**
     * Encode calls as RLP tuples for atomic batch execution
     * @returns Array of [to, value, data] tuples
     * @private
     */
    encodeCallsAsTuples() {
        return this.txRequest.calls.map((call) => [call.to, this.bigintToHex(call.value), call.data]);
    }
    /**
     * Encode EIP-2930 access list as RLP tuples
     * @returns Array of [address, storageKeys[]] tuples
     * @private
     */
    encodeAccessList() {
        return (this.txRequest.accessList ?? []).map((item) => [item.address, item.storageKeys || []]);
    }
    /**
     * Convert bigint to hex string for RLP encoding
     * @param value bigint value to convert
     * @returns Hex string
     * @private
     */
    bigintToHex(value) {
        if (value === 0n) {
            return '0x';
        }
        const hex = value.toString(16);
        return '0x' + (hex.length % 2 ? '0' : '') + hex;
    }
    /**
     * Build base RLP data array per Tempo EIP-7702 specification
     * @param callsTuples Encoded calls
     * @param accessTuples Encoded access list
     * @returns RLP-ready array of transaction fields
     * @private
     */
    buildBaseRlpData(callsTuples, accessTuples) {
        return [
            ethers_1.ethers.utils.hexlify(this.txRequest.chainId),
            this.txRequest.maxPriorityFeePerGas ? this.bigintToHex(this.txRequest.maxPriorityFeePerGas) : '0x',
            this.bigintToHex(this.txRequest.maxFeePerGas),
            this.bigintToHex(this.txRequest.gas),
            callsTuples,
            accessTuples,
            '0x', // nonceKey (reserved for 2D nonce system)
            this.txRequest.nonce ? this.bigintToHex(BigInt(this.txRequest.nonce)) : '0x',
            '0x', // validBefore (reserved for time bounds)
            '0x', // validAfter (reserved for time bounds)
            this.txRequest.feeToken || '0x',
            '0x', // feePayerSignature (reserved for sponsorship)
            [], // authorizationList (EIP-7702)
        ];
    }
    /**
     * Encode secp256k1 signature as 65-byte envelope
     * @param signature ECDSA signature components
     * @returns Hex string of concatenated r (32) + s (32) + v (1) bytes
     * @private
     */
    encodeSignature(signature) {
        const v = signature.yParity + 27;
        const signatureBytes = ethers_1.ethers.utils.concat([
            ethers_1.ethers.utils.zeroPad(signature.r, 32),
            ethers_1.ethers.utils.zeroPad(signature.s, 32),
            ethers_1.ethers.utils.hexlify(v),
        ]);
        return ethers_1.ethers.utils.hexlify(signatureBytes);
    }
    /**
     * RLP encode and prepend transaction type byte
     * @param rlpData Transaction fields array
     * @returns Hex string with 0x76 prefix
     * @private
     */
    rlpEncodeWithTypePrefix(rlpData) {
        try {
            const encoded = ethers_1.ethers.utils.RLP.encode(rlpData);
            return ('0x76' + encoded.slice(2));
        }
        catch (error) {
            throw new sdk_core_1.ParseTransactionError(`Failed to RLP encode transaction: ${error}`);
        }
    }
    /**
     * Serialize Tempo AA transaction (type 0x76) per EIP-7702 specification
     * Format: 0x76 || RLP([chainId, fees, gas, calls, accessList, nonce fields, feeToken, sponsorship, authList, signature?])
     * @param signature Optional ECDSA signature (omit for unsigned transactions)
     * @returns RLP-encoded transaction hex string
     * @private
     */
    serializeTransaction(signature) {
        const callsTuples = this.encodeCallsAsTuples();
        const accessTuples = this.encodeAccessList();
        const rlpData = this.buildBaseRlpData(callsTuples, accessTuples);
        if (signature) {
            rlpData.push(this.encodeSignature(signature));
        }
        return this.rlpEncodeWithTypePrefix(rlpData);
    }
    getOperations() {
        return [...this._operations];
    }
    getFeeToken() {
        return this.txRequest.feeToken;
    }
    getOperationCount() {
        return this.txRequest.calls.length;
    }
    isBatch() {
        return this.txRequest.calls.length > 1;
    }
    setSignature(signature) {
        this._signature = signature;
    }
    getSignature() {
        return this._signature;
    }
    /**
     * Recover the signer address from the transaction signature.
     * @returns Checksummed sender address, or undefined if unsigned
     */
    getSenderAddress() {
        if (!this._signature) {
            return undefined;
        }
        const unsignedHex = this.serializeTransaction();
        const msgHash = ethers_1.ethers.utils.keccak256(ethers_1.ethers.utils.arrayify(unsignedHex));
        return ethers_1.ethers.utils.recoverAddress(msgHash, {
            r: this._signature.r,
            s: this._signature.s,
            recoveryParam: this._signature.yParity,
        });
    }
    toJson() {
        return {
            type: this.txRequest.type,
            chainId: this.txRequest.chainId,
            nonce: this.txRequest.nonce,
            maxFeePerGas: this.txRequest.maxFeePerGas.toString(),
            maxPriorityFeePerGas: this.txRequest.maxPriorityFeePerGas.toString(),
            gas: this.txRequest.gas.toString(),
            callCount: this.txRequest.calls.length,
            feeToken: this.txRequest.feeToken,
            operations: this._operations,
            signature: this._signature,
        };
    }
    async toBroadcastFormat() {
        return await this.serialize(this._signature);
    }
    /** @inheritdoc */
    get id() {
        try {
            const serialized = this.serializeTransaction(this._signature);
            return ethers_1.ethers.utils.keccak256(ethers_1.ethers.utils.arrayify(serialized));
        }
        catch {
            return 'pending';
        }
    }
    toString() {
        return JSON.stringify(this.toJson(), null, 2);
    }
    canBroadcast() {
        return this.txRequest.calls.length > 0 && this.txRequest.chainId > 0;
    }
}
exports.Tip20Transaction = Tip20Transaction;
