// @bitgo/sdk-coin-tempo/dist/src/lib/utils.js
"use strict";
/**
 * Tempo Utility Functions
 *
 * Since Tempo is EVM-compatible, we can reuse Ethereum utilities
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidAddress = isValidAddress;
exports.isValidPublicKey = isValidPublicKey;
exports.isValidPrivateKey = isValidPrivateKey;
exports.amountToTip20Units = amountToTip20Units;
exports.tip20UnitsToAmount = tip20UnitsToAmount;
exports.stringToBytes32 = stringToBytes32;
exports.encodeTip20TransferWithMemo = encodeTip20TransferWithMemo;
exports.isValidTip20Amount = isValidTip20Amount;
exports.isTip20Transaction = isTip20Transaction;
exports.isValidMemoId = isValidMemoId;
const secp256k1_1 = require("@bitgo/secp256k1");
const ethers_1 = require("ethers");
const constants_1 = require("./constants");
const tip20Abi_1 = require("./tip20Abi");
const AA_TX_HEX_REGEX = new RegExp(`^${constants_1.AA_TRANSACTION_TYPE}[0-9a-f]*$`, 'i');
/**
 * Check if address is valid Ethereum-style address
 * Uses ethers.js isAddress for proper validation including checksum
 */
function isValidAddress(address) {
    if (typeof address !== 'string') {
        return false;
    }
    return ethers_1.ethers.utils.isAddress(address);
}
/**
 * Check if public key is valid (BIP32 xpub format)
 * TODO: Replace with ETH utils when implementing
 */
function isValidPublicKey(publicKey) {
    if (typeof publicKey !== 'string') {
        return false;
    }
    try {
        const hdNode = secp256k1_1.bip32.fromBase58(publicKey);
        return hdNode.isNeutered();
    }
    catch (e) {
        return false;
    }
}
/**
 * Check if private key is valid (BIP32 xprv format)
 * TODO: Replace with ETH utils when implementing
 */
function isValidPrivateKey(privateKey) {
    if (typeof privateKey !== 'string') {
        return false;
    }
    try {
        const hdNode = secp256k1_1.bip32.fromBase58(privateKey);
        return !hdNode.isNeutered();
    }
    catch (e) {
        return false;
    }
}
/**
 * TIP-20 Utility Functions
 */
/**
 * Convert human-readable amount to TIP-20 units (6 decimals)
 * @param amount - Human-readable amount (e.g., "1.5")
 * @returns Amount in TIP-20 smallest units as bigint
 * @example amountToTip20Units("1.5") => 1500000n
 */
function amountToTip20Units(amount) {
    try {
        return BigInt(ethers_1.ethers.utils.parseUnits(amount, constants_1.TIP20_DECIMALS).toString());
    }
    catch (error) {
        throw new Error(`Invalid amount format: ${amount}. Expected decimal string.`);
    }
}
/**
 * Convert TIP-20 units (6 decimals) to human-readable amount
 * @param units - Amount in TIP-20 smallest units
 * @returns Human-readable amount string
 * @example tip20UnitsToAmount(1500000n) => "1.5"
 */
function tip20UnitsToAmount(units) {
    return ethers_1.ethers.utils.formatUnits(units.toString(), constants_1.TIP20_DECIMALS);
}
/**
 * Convert string to bytes32 for memo field
 * @param memo - Memo string to encode
 * @returns Hex-encoded bytes32 value
 * @example stringToBytes32("INVOICE-001") => "0x494e564f4943452d30303100..."
 */
function stringToBytes32(memo) {
    const memoByteLength = new TextEncoder().encode(memo).length;
    if (memoByteLength > 32) {
        throw new Error(`Memo too long: ${memoByteLength} bytes. Maximum 32 bytes.`);
    }
    const hexString = ethers_1.ethers.utils.hexlify(ethers_1.ethers.utils.toUtf8Bytes(memo));
    return ethers_1.ethers.utils.hexZeroPad(hexString, 32);
}
/**
 * Encode TIP-20 transferWithMemo function call using ethers.js
 * @param to - Recipient address
 * @param amount - Amount in TIP-20 units (bigint)
 * @param memo - Optional memo string
 * @returns Encoded function call data
 */
function encodeTip20TransferWithMemo(to, amount, memo) {
    const memoBytes = memo ? stringToBytes32(memo) : ethers_1.ethers.utils.hexZeroPad('0x', 32);
    const iface = new ethers_1.ethers.utils.Interface(tip20Abi_1.TIP20_TRANSFER_WITH_MEMO_ABI);
    return iface.encodeFunctionData('transferWithMemo', [to, amount, memoBytes]);
}
/**
 * Validate TIP-20 amount format
 * @param amount - Amount string to validate
 * @returns true if valid, false otherwise
 */
function isValidTip20Amount(amount) {
    if (typeof amount !== 'string' || amount.trim() === '') {
        return false;
    }
    // Check for negative amounts before parsing
    if (amount.startsWith('-')) {
        return false;
    }
    try {
        const parsed = ethers_1.ethers.utils.parseUnits(amount, constants_1.TIP20_DECIMALS);
        return parsed.gte(0);
    }
    catch {
        return false;
    }
}
/**
 * Check if a raw transaction string is a Tempo AA transaction (type 0x76)
 */
function isTip20Transaction(raw) {
    return AA_TX_HEX_REGEX.test(raw);
}
/**
 * Validate that a memoId is a valid non-negative integer string
 */
function isValidMemoId(memoId) {
    return typeof memoId === 'string' && /^(0|[1-9]\d*)$/.test(memoId);
}
const utils = {
    isValidAddress,
    isValidPublicKey,
    isValidPrivateKey,
    amountToTip20Units,
    tip20UnitsToAmount,
    stringToBytes32,
    encodeTip20TransferWithMemo,
    isValidTip20Amount,
    isTip20Transaction,
    isValidMemoId,
};
exports.default = utils;
