// @bitgo/sdk-coin-xrp/dist/src/ripple.js
"use strict";
/**
 * @hidden
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
/**
 */
const rippleKeypairs = __importStar(require("ripple-keypairs"));
const xrpl = __importStar(require("xrpl"));
const secp256k1_1 = require("@bitgo/secp256k1");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const binary = __importStar(require("ripple-binary-codec"));
/**
 * Convert an XRP address to a BigNumber for numeric comparison.
 * This is needed for proper sorting of signers as required by the XRP protocol.
 *
 * @param address - The XRP address to convert
 * @returns BigNumber representation of the address
 */
function addressToBigNumber(address) {
    const hex = Buffer.from(xrpl.decodeAccountID(address)).toString('hex');
    return new bignumber_js_1.default(hex, 16);
}
function computeSignature(tx, privateKey, signAs) {
    const signingData = signAs ? binary.encodeForMultisigning(tx, signAs) : binary.encodeForSigning(tx);
    return rippleKeypairs.sign(signingData, privateKey);
}
/**
 * Sign Ripple transaction with a secp256k1 private key
 * @param txHex
 * @param privateKey
 * @param options
 * @returns {{signedTransaction: *, id}}
 */
const signWithPrivateKey = function (txHex, privateKey, options) {
    let privateKeyBuffer = Buffer.from(privateKey, 'hex');
    if (privateKeyBuffer.length === 33 && privateKeyBuffer[0] === 0) {
        privateKeyBuffer = privateKeyBuffer.slice(1, 33);
    }
    const publicKey = secp256k1_1.ECPair.fromPrivateKey(privateKeyBuffer).publicKey.toString('hex').toUpperCase();
    let tx;
    try {
        tx = binary.decode(txHex);
    }
    catch (e) {
        try {
            tx = JSON.parse(txHex);
        }
        catch (e) {
            throw new Error('txHex needs to be either hex or JSON string for XRP');
        }
    }
    tx.SigningPubKey = options && options.signAs ? '' : publicKey;
    if (options && options.signAs) {
        const expectedSigner = rippleKeypairs.deriveAddress(publicKey);
        if (options.signAs !== expectedSigner) {
            throw new Error('signAs does not match private key');
        }
        const signer = {
            Account: options.signAs,
            SigningPubKey: publicKey,
            TxnSignature: computeSignature(tx, privateKey, options.signAs),
        };
        // Ordering of private key signing matters, or the Ripple fullnode will throw an 'Unsorted Signers array' error.
        // XRP requires Signers array to be sorted based on numeric value of signer addresses (lowest first)
        if (tx.Signers) {
            // Add the current signer
            tx.Signers.push({ Signer: signer });
            // Sort the Signers array by numeric value of Account (address) to ensure proper ordering
            tx.Signers.sort((a, b) => {
                const addressBN1 = addressToBigNumber(a.Signer.Account);
                const addressBN2 = addressToBigNumber(b.Signer.Account);
                return addressBN1.comparedTo(addressBN2);
            });
        }
        else {
            tx.Signers = [{ Signer: signer }];
        }
    }
    else {
        tx.TxnSignature = computeSignature(tx, privateKey, undefined);
    }
    const serialized = binary.encode(tx);
    return {
        signedTransaction: serialized,
        id: xrpl.hashes.hashSignedTx(serialized),
    };
};
module.exports = { ...xrpl, signWithPrivateKey };
