// @bitgo/sdk-coin-xrp/dist/src/xrp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Xrp = void 0;
/**
 * @prettier
 */
const bignumber_js_1 = require("bignumber.js");
const _ = __importStar(require("lodash"));
const querystring = __importStar(require("querystring"));
const url = __importStar(require("url"));
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const rippleBinaryCodec = __importStar(require("ripple-binary-codec"));
const rippleKeypairs = __importStar(require("ripple-keypairs"));
const xrpl = __importStar(require("xrpl"));
const lib_1 = require("./lib");
const iface_1 = require("./lib/iface");
const keyPair_1 = require("./lib/keyPair");
const utils_1 = __importDefault(require("./lib/utils"));
const ripple_1 = __importDefault(require("./ripple"));
class Xrp extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Xrp(bitgo, staticsCoin);
    }
    /**
     * Factor between the coin's base unit and its smallest subdivison
     */
    getBaseFactor() {
        return Math.pow(10, this._staticsCoin.decimalPlaces);
    }
    /**
     * Identifier for the blockchain which supports this coin
     */
    getChain() {
        return this._staticsCoin.name;
    }
    /**
     * Identifier for the coin family
     */
    getFamily() {
        return this._staticsCoin.family;
    }
    /**
     * Complete human-readable name of this coin
     */
    getFullName() {
        return this._staticsCoin.fullName;
    }
    /**
     * Evaluates whether an address string is valid for this coin
     * @param address
     */
    isValidAddress(address) {
        return utils_1.default.isValidAddress(address);
    }
    /**
     * Return boolean indicating whether input is valid public key for the coin.
     *
     * @param {String} pub the pub to be checked
     * @returns {Boolean} is it valid?
     */
    isValidPub(pub) {
        return utils_1.default.isValidPublicKey(pub);
    }
    /**
     * Get fee info from server
     */
    async getFeeInfo() {
        return this.bitgo.get(this.url('/public/feeinfo')).result();
    }
    /** @inheritdoc */
    valuelessTransferAllowed() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.onchain;
    }
    getTokenEnablementConfig() {
        return {
            requiresTokenEnablement: true,
            supportsMultipleTokenEnablements: false,
        };
    }
    /**
     * Assemble keychain and half-sign prebuilt transaction
     * @param params
     * - txPrebuild
     * - prv
     * @returns Bluebird<HalfSignedTransaction>
     */
    async signTransaction({ txPrebuild, prv, isLastSignature, }) {
        if (_.isUndefined(txPrebuild) || !_.isObject(txPrebuild)) {
            if (!_.isUndefined(txPrebuild) && !_.isObject(txPrebuild)) {
                throw new Error(`txPrebuild must be an object, got type ${typeof txPrebuild}`);
            }
            throw new Error('missing txPrebuild parameter');
        }
        if (_.isUndefined(prv) || !_.isString(prv)) {
            if (!_.isUndefined(prv) && !_.isString(prv)) {
                throw new Error(`prv must be a string, got type ${typeof prv}`);
            }
            throw new Error('missing prv parameter to sign transaction');
        }
        if (!txPrebuild.txHex) {
            throw new Error(`missing txHex in txPrebuild`);
        }
        const keyPair = new keyPair_1.KeyPair({ prv });
        const address = keyPair.getAddress();
        const privateKey = keyPair.getPrivateKey().toString('hex');
        const tx = ripple_1.default.signWithPrivateKey(txPrebuild.txHex, privateKey, {
            signAs: address,
        });
        // Normally the SDK provides the first signature for an XRP tx, but occasionally it provides the final one as well
        // (recoveries)
        if (isLastSignature) {
            return { txHex: tx.signedTransaction };
        }
        return { halfSigned: { txHex: tx.signedTransaction } };
    }
    /**
     * Ripple requires additional parameters for wallet generation to be sent to the server. The additional parameters are
     * the root public key, which is the basis of the root address, two signed, and one half-signed initialization txs
     * @param walletParams
     * - rootPrivateKey: optional hex-encoded Ripple private key
     */
    async supplementGenerateWallet(walletParams) {
        if (walletParams.rootPrivateKey) {
            if (walletParams.rootPrivateKey.length !== 64) {
                throw new Error('rootPrivateKey needs to be a hexadecimal private key string');
            }
        }
        else {
            const keyPair = new keyPair_1.KeyPair().getKeys();
            if (!keyPair.prv) {
                throw new Error('no privateKey');
            }
            walletParams.rootPrivateKey = keyPair.prv;
        }
        return walletParams;
    }
    /**
     * Explain/parse transaction
     * @param params
     */
    async explainTransaction(params = {}) {
        let transaction;
        let txHex = params.txHex || (params.halfSigned && params.halfSigned.txHex);
        if (!txHex) {
            throw new Error('missing required param txHex');
        }
        try {
            transaction = rippleBinaryCodec.decode(txHex);
        }
        catch (e) {
            try {
                transaction = JSON.parse(txHex);
                txHex = rippleBinaryCodec.encode(transaction);
            }
            catch (e) {
                throw new Error('txHex needs to be either hex or JSON string for XRP');
            }
        }
        let id;
        // hashes ids are different for signed and unsigned tx
        // first we try to get the hash id as if it is signed, will throw if its not
        try {
            id = xrpl.hashes.hashSignedTx(txHex);
        }
        catch (e) {
            id = xrpl.hashes.hashTx(txHex);
        }
        if (transaction.TransactionType === 'AccountSet') {
            return {
                displayOrder: ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'accountSet'],
                id: id,
                changeOutputs: [],
                outputAmount: 0,
                changeAmount: 0,
                outputs: [],
                fee: {
                    fee: transaction.Fee,
                    feeRate: undefined,
                    size: txHex.length / 2,
                },
                accountSet: {
                    messageKey: transaction.MessageKey,
                    setFlag: transaction.SetFlag,
                },
            };
        }
        else if (transaction.TransactionType === 'TrustSet') {
            return {
                displayOrder: [
                    'id',
                    'outputAmount',
                    'changeAmount',
                    'outputs',
                    'changeOutputs',
                    'fee',
                    'account',
                    'limitAmount',
                ],
                id: id,
                changeOutputs: [],
                outputAmount: 0,
                changeAmount: 0,
                outputs: [],
                fee: {
                    fee: transaction.Fee,
                    feeRate: undefined,
                    size: txHex.length / 2,
                },
                account: transaction.Account,
                limitAmount: {
                    currency: transaction.LimitAmount.currency,
                    issuer: transaction.LimitAmount.issuer,
                    value: transaction.LimitAmount.value,
                },
            };
        }
        const address = transaction.Destination + (transaction.DestinationTag >= 0 ? '?dt=' + transaction.DestinationTag : '');
        return {
            displayOrder: ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee'],
            id: id,
            changeOutputs: [],
            outputAmount: transaction.Amount,
            changeAmount: 0,
            outputs: [
                {
                    address,
                    amount: transaction.Amount,
                },
            ],
            fee: {
                fee: transaction.Fee,
                feeRate: undefined,
                size: txHex.length / 2,
            },
        };
    }
    getTransactionTypeRawTxHex(txHex) {
        let transaction;
        if (!txHex) {
            throw new Error('missing required param txHex');
        }
        try {
            transaction = rippleBinaryCodec.decode(txHex);
        }
        catch (e) {
            try {
                transaction = JSON.parse(txHex);
            }
            catch (e) {
                throw new Error('txHex needs to be either hex or JSON string for XRP');
            }
        }
        return transaction.TransactionType;
    }
    verifyTxType(txPrebuildDecoded, txHexPrebuild) {
        if (!txHexPrebuild)
            throw new Error('Missing txHexPrebuild to verify token type for enabletoken tx');
        const transactionType = this.getTransactionTypeRawTxHex(txHexPrebuild);
        if (transactionType === undefined)
            throw new Error('Missing TransactionType on token enablement tx');
        if (transactionType !== iface_1.XrpTransactionType.TrustSet)
            throw new Error(`tx type ${transactionType} does not match expected type TrustSet`);
        // decoded payload type could come as undefined or any of the enabletoken like types but never as something else like Send, etc
        const actualTypeFromDecoded = 'type' in txPrebuildDecoded && typeof txPrebuildDecoded.type === 'string' ? txPrebuildDecoded.type : undefined;
        if (!actualTypeFromDecoded ||
            actualTypeFromDecoded === 'enabletoken' ||
            actualTypeFromDecoded === 'AssociatedTokenAccountInitialization')
            return;
        throw new Error(`tx type ${actualTypeFromDecoded} does not match the expected type enabletoken`);
    }
    verifyTokenName(txParams, txPrebuildDecoded, txHexPrebuild, coinConfig) {
        if (!txHexPrebuild)
            throw new Error('Missing txHexPrebuild param required for token enablement.');
        if (!txParams.recipients || txParams.recipients.length === 0)
            throw new Error('Missing recipients param for token enablement.');
        const fullTokenName = txParams.recipients[0].tokenName;
        if (fullTokenName === undefined)
            throw new Error('Param tokenName is required for token enablement. Recipient must include a token name.');
        if (!('limitAmount' in txPrebuildDecoded))
            throw new Error('Missing limitAmount param for token enablement.');
        // we check currency on both the txHex but also the explained payload
        const expectedCurrency = utils_1.default.getXrpCurrencyFromTokenName(fullTokenName).currency;
        if (coinConfig.isToken && expectedCurrency !== txPrebuildDecoded.limitAmount.currency)
            throw new Error('Invalid token issuer or currency on token enablement tx');
    }
    verifyActivationAddress(txParams, txPrebuildDecoded) {
        if (txParams.recipients === undefined || txParams.recipients.length === 0)
            throw new Error('Missing recipients param for token enablement.');
        if (txParams.recipients?.length !== 1) {
            throw new Error(`${this.getChain()} doesn't support sending to more than 1 destination address within a single transaction. Try again, using only a single recipient.`);
        }
        if (!('account' in txPrebuildDecoded))
            throw new Error('missing account on token enablement tx');
        const activationAddress = txParams.recipients[0].address;
        const accountAddress = txPrebuildDecoded.account;
        if (activationAddress !== accountAddress)
            throw new Error("Account address doesn't match with activation address.");
    }
    verifyTokenIssuer(txParams, txPrebuildDecoded) {
        if (txPrebuildDecoded === undefined || !('limitAmount' in txPrebuildDecoded))
            throw new Error('missing token issuer on token enablement tx');
        const { issuer, currency } = txPrebuildDecoded.limitAmount;
        if (!utils_1.default.getXrpToken(issuer, currency))
            throw new Error('Invalid token issuer or currency on token enablement tx');
    }
    verifyRequiredKeys(txParams, txPrebuildDecoded) {
        if (!('account' in txPrebuildDecoded) ||
            !('limitAmount' in txPrebuildDecoded) ||
            !txPrebuildDecoded.limitAmount.currency) {
            throw new Error('Explanation is missing required keys (account or limitAmount with currency)');
        }
    }
    /**
     * Verify that a transaction prebuild complies with the original intention
     * @param txParams params object passed to send
     * @param txPrebuild prebuild object returned by server
     * @param wallet
     * @returns {boolean}
     */
    async verifyTransaction({ txParams, txPrebuild, verification }) {
        const coinConfig = statics_1.coins.get(this.getChain());
        const explanation = await this.explainTransaction({
            txHex: txPrebuild.txHex,
        });
        // Explaining a tx strips out certain data, for extra measurement we're checking vs the explained tx
        // but also vs the tx pre explained.
        if (txParams.type === 'enabletoken' && verification?.verifyTokenEnablement) {
            this.verifyTxType(explanation, txPrebuild.txHex);
            this.verifyActivationAddress(txParams, explanation);
            this.verifyTokenIssuer(txParams, explanation);
            this.verifyTokenName(txParams, explanation, txPrebuild.txHex, coinConfig);
            this.verifyRequiredKeys(txParams, explanation);
        }
        const output = [...explanation.outputs, ...explanation.changeOutputs][0];
        const expectedOutput = txParams.recipients && txParams.recipients[0];
        const comparator = (recipient1, recipient2) => {
            if (utils_1.default.getAddressDetails(recipient1.address).address !== utils_1.default.getAddressDetails(recipient2.address).address) {
                return false;
            }
            const amount1 = new bignumber_js_1.BigNumber(recipient1.amount);
            const amount2 = new bignumber_js_1.BigNumber(recipient2.amount);
            return amount1.toFixed() === amount2.toFixed();
        };
        if ((txParams.type === undefined || txParams.type === 'payment') &&
            typeof output.amount !== 'object' &&
            !comparator(output, expectedOutput)) {
            throw new Error('transaction prebuild does not match expected output');
        }
        return true;
    }
    /**
     * Check if address is a valid XRP address, and then make sure the root addresses match.
     * This prevents attacks where an attack may switch out the new address for one of their own
     * @param address {String} the address to verify
     * @param rootAddress {String} the wallet's root address
     * @return true iff address is a wallet address (based on rootAddress)
     */
    async isWalletAddress({ address, rootAddress }) {
        if (!this.isValidAddress(address)) {
            throw new sdk_core_1.InvalidAddressError(`address verification failure: address "${address}" is not valid`);
        }
        const accountInfoParams = {
            method: 'account_info',
            params: [
                {
                    account: address,
                    ledger_index: 'current',
                    queue: true,
                    strict: true,
                    signer_lists: true,
                },
            ],
        };
        const accountInfo = (await this.bitgo.post(this.getRippledUrl()).send(accountInfoParams)).body;
        if (accountInfo?.result?.account_data?.Flags == null) {
            throw new Error('Invalid account information: Flags field is missing.');
        }
        const flags = xrpl.parseAccountRootFlags(accountInfo.result.account_data.Flags);
        const addressDetails = utils_1.default.getAddressDetails(address);
        const rootAddressDetails = utils_1.default.getAddressDetails(rootAddress);
        if (flags.lsfRequireDestTag && addressDetails.destinationTag == null) {
            throw new sdk_core_1.InvalidAddressError(`Invalid Address: Destination Tag is required for address "${address}".`);
        }
        if (addressDetails.address !== rootAddressDetails.address) {
            throw new sdk_core_1.UnexpectedAddressError(`address validation failure: ${addressDetails.address} vs. ${rootAddressDetails.address}`);
        }
        return true;
    }
    /**
     * URL of a well-known, public facing (non-bitgo) rippled instance which can be used for recovery
     */
    getRippledUrl() {
        return 'https://s1.ripple.com:51234';
    }
    /**
     * Builds a funds recovery transaction without BitGo
     * @param params
     * - rootAddress: root XRP wallet address to recover funds from
     * - userKey: [encrypted] xprv
     * - backupKey: [encrypted] xprv, or xpub if the xprv is held by a KRS provider
     * - walletPassphrase: necessary if one of the xprvs is encrypted
     * - bitgoKey: xpub
     * - krsProvider: necessary if backup key is held by KRS
     * - recoveryDestination: target address to send recovered funds to
     */
    async recover(params) {
        const rippledUrl = this.getRippledUrl();
        const isKrsRecovery = params.backupKey.startsWith('xpub') && !params.userKey.startsWith('xpub');
        const isUnsignedSweep = params.backupKey.startsWith('xpub') && params.userKey.startsWith('xpub');
        const accountInfoParams = {
            method: 'account_info',
            params: [
                {
                    account: params.rootAddress,
                    ledger_index: 'current',
                    queue: true,
                    strict: true,
                    signer_lists: true,
                },
            ],
        };
        const accountLinesParams = {
            method: 'account_lines',
            params: [
                {
                    account: params.rootAddress,
                    ledger_index: 'validated',
                },
            ],
        };
        if (isKrsRecovery) {
            (0, sdk_core_1.checkKrsProvider)(this, params.krsProvider);
        }
        // Validate the destination address
        if (!this.isValidAddress(params.recoveryDestination)) {
            throw new Error('Invalid destination address!');
        }
        const keys = (0, sdk_core_1.getBip32Keys)(this.bitgo, params, { requireBitGoXpub: false });
        const { addressDetails, feeDetails, serverDetails, accountLines } = await (0, sdk_core_1.promiseProps)({
            addressDetails: this.bitgo.post(rippledUrl).send(accountInfoParams),
            feeDetails: this.bitgo.post(rippledUrl).send({ method: 'fee' }),
            serverDetails: this.bitgo.post(rippledUrl).send({ method: 'server_info' }),
            accountLines: this.bitgo.post(rippledUrl).send(accountLinesParams),
        });
        const openLedgerFee = new bignumber_js_1.BigNumber(feeDetails.body.result.drops.open_ledger_fee);
        const baseReserve = new bignumber_js_1.BigNumber(serverDetails.body.result.info.validated_ledger.reserve_base_xrp).times(this.getBaseFactor());
        const reserveDelta = new bignumber_js_1.BigNumber(serverDetails.body.result.info.validated_ledger.reserve_inc_xrp).times(this.getBaseFactor());
        const currentLedger = serverDetails.body.result.info.validated_ledger.seq;
        const sequenceId = addressDetails.body.result.account_data.Sequence;
        const balance = new bignumber_js_1.BigNumber(addressDetails.body.result.account_data.Balance);
        const signerLists = addressDetails.body.result.account_data.signer_lists;
        const accountFlags = addressDetails.body.result.account_data.Flags;
        const ownerCount = new bignumber_js_1.BigNumber(addressDetails.body.result.account_data.OwnerCount);
        // make sure there is only one signer list set
        if (signerLists.length !== 1) {
            throw new Error('unexpected set of signer lists');
        }
        // make sure the signers are user, backup, bitgo
        const userAddress = rippleKeypairs.deriveAddress(keys[0].publicKey.toString('hex'));
        const backupAddress = rippleKeypairs.deriveAddress(keys[1].publicKey.toString('hex'));
        const signerList = signerLists[0];
        if (signerList.SignerQuorum !== 2) {
            throw new Error('invalid minimum signature count');
        }
        const foundAddresses = {};
        const signerEntries = signerList.SignerEntries;
        if (signerEntries.length !== 3) {
            throw new Error('invalid signer list length');
        }
        for (const { SignerEntry } of signerEntries) {
            const weight = SignerEntry.SignerWeight;
            const address = SignerEntry.Account;
            if (weight !== 1) {
                throw new Error('invalid signer weight');
            }
            // if it's a dupe of an address we already know, block
            if (foundAddresses[address] >= 1) {
                throw new Error('duplicate signer address');
            }
            foundAddresses[address] = (foundAddresses[address] || 0) + 1;
        }
        if (foundAddresses[userAddress] !== 1) {
            throw new Error('unexpected incidence frequency of user signer address');
        }
        if (foundAddresses[backupAddress] !== 1) {
            throw new Error('unexpected incidence frequency of user signer address');
        }
        // make sure the flags disable the master key and enforce destination tags
        const USER_KEY_SETTING_FLAG = 65536;
        const MASTER_KEY_DEACTIVATION_FLAG = 1048576;
        const REQUIRE_DESTINATION_TAG_FLAG = 131072;
        if ((accountFlags & USER_KEY_SETTING_FLAG) !== 0) {
            throw new Error('a custom user key has been set');
        }
        if ((accountFlags & MASTER_KEY_DEACTIVATION_FLAG) !== MASTER_KEY_DEACTIVATION_FLAG) {
            throw new Error('the master key has not been deactivated');
        }
        if ((accountFlags & REQUIRE_DESTINATION_TAG_FLAG) !== REQUIRE_DESTINATION_TAG_FLAG) {
            throw new Error('the destination flag requirement has not been activated');
        }
        // recover the funds
        const totalReserveDelta = reserveDelta.times(ownerCount);
        const reserve = baseReserve.plus(totalReserveDelta);
        const recoverableBalance = balance.minus(reserve);
        const rawDestination = params.recoveryDestination;
        const destinationDetails = url.parse(rawDestination);
        if (destinationDetails.query) {
            const queryDetails = querystring.parse(destinationDetails.query);
            if (Array.isArray(queryDetails.dt)) {
                // if queryDetails.dt is an array, that means dt was given multiple times, which is not valid
                throw new sdk_core_1.InvalidAddressError(`destination tag can appear at most once, but ${queryDetails.dt.length} destination tags were found`);
            }
        }
        if (recoverableBalance.toNumber() <= 0) {
            throw new Error(`Quantity of XRP to recover must be greater than 0. Current balance: ${balance.toNumber()}, blockchain reserve: ${reserve.toNumber()}, spendable balance: ${recoverableBalance.toNumber()}`);
        }
        const issuer = params?.issuerAddress;
        const currency = params?.currencyCode;
        if (!!issuer && !!currency) {
            const tokenParams = {
                recoveryDestination: params.recoveryDestination,
                recoverableBalance,
                currentLedger,
                openLedgerFee,
                sequenceId,
                accountLines,
                keys,
                isKrsRecovery,
                isUnsignedSweep,
                userAddress,
                backupAddress,
                issuer,
                currency,
            };
            return this.recoverXrpToken(params, tokenParams);
        }
        const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
        const txBuilder = factory.getTransferBuilder();
        txBuilder
            .to(params.recoveryDestination)
            .amount(recoverableBalance.toFixed(0))
            .sender(params.rootAddress)
            .flags(2147483648)
            .lastLedgerSequence(currentLedger + 1000000) // give it 1 million ledgers' time (~1 month, suitable for KRS)
            .fee(openLedgerFee.times(3).toFixed(0)) // the factor three is for the multisigning
            .sequence(sequenceId);
        const tx = await txBuilder.build();
        const serializedTx = tx.toBroadcastFormat();
        if (isUnsignedSweep) {
            return {
                txHex: serializedTx,
                coin: this.getChain(),
            };
        }
        if (!keys[0].privateKey) {
            throw new Error(`userKey is not a private key`);
        }
        const userKey = keys[0].privateKey.toString('hex');
        const userSignature = ripple_1.default.signWithPrivateKey(serializedTx, userKey, { signAs: userAddress });
        let signedTransaction;
        if (isKrsRecovery) {
            signedTransaction = userSignature.signedTransaction;
        }
        else {
            if (!keys[1].privateKey) {
                throw new Error(`backupKey is not a private key`);
            }
            const backupKey = keys[1].privateKey.toString('hex');
            const backupSignature = ripple_1.default.signWithPrivateKey(serializedTx, backupKey, { signAs: backupAddress });
            signedTransaction = ripple_1.default.multisign([userSignature.signedTransaction, backupSignature.signedTransaction]);
        }
        const transactionExplanation = (await this.explainTransaction({
            txHex: signedTransaction,
        }));
        transactionExplanation.txHex = signedTransaction;
        if (isKrsRecovery) {
            transactionExplanation.backupKey = params.backupKey;
            transactionExplanation.coin = this.getChain();
        }
        return transactionExplanation;
    }
    async recoverXrpToken(params, tokenParams) {
        const { currency, issuer } = tokenParams;
        const tokenName = utils_1.default.getXrpToken(issuer, currency).name;
        const lines = tokenParams.accountLines.body.result.lines;
        let amount;
        for (const line of lines) {
            if (line.currency === currency && line.account === issuer) {
                amount = line.balance;
                break;
            }
        }
        if (amount === undefined) {
            throw new Error(`Does not have Trustline with ${issuer}`);
        }
        if (amount === '0') {
            throw new Error(`Does not have funds to recover`);
        }
        const decimalPlaces = statics_1.coins.get(tokenName).decimalPlaces;
        amount = new bignumber_js_1.BigNumber(amount).shiftedBy(decimalPlaces).toFixed();
        const FLAG_VALUE = 2147483648;
        const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get(tokenName));
        const txBuilder = factory.getTokenTransferBuilder();
        txBuilder
            .to(tokenParams.recoveryDestination)
            .amount(amount)
            .sender(params.rootAddress)
            .flags(FLAG_VALUE)
            .lastLedgerSequence(tokenParams.currentLedger + 1000000) // give it 1 million ledgers' time (~1 month, suitable for KRS)
            .fee(tokenParams.openLedgerFee.times(3).toFixed(0)) // the factor three is for the multisigning
            .sequence(tokenParams.sequenceId);
        const tx = await txBuilder.build();
        const serializedTx = tx.toBroadcastFormat();
        const { keys, isKrsRecovery, isUnsignedSweep, userAddress, backupAddress } = tokenParams;
        if (isUnsignedSweep) {
            return {
                txHex: serializedTx,
                coin: this.getChain(),
            };
        }
        if (!keys[0].privateKey) {
            throw new Error(`userKey is not a private key`);
        }
        const userKey = keys[0].privateKey.toString('hex');
        const userSignature = ripple_1.default.signWithPrivateKey(serializedTx, userKey, { signAs: userAddress });
        let signedTransaction;
        if (isKrsRecovery) {
            signedTransaction = userSignature.signedTransaction;
        }
        else {
            if (!keys[1].privateKey) {
                throw new Error(`backupKey is not a private key`);
            }
            const backupKey = keys[1].privateKey.toString('hex');
            const backupSignature = ripple_1.default.signWithPrivateKey(serializedTx, backupKey, { signAs: backupAddress });
            signedTransaction = ripple_1.default.multisign([userSignature.signedTransaction, backupSignature.signedTransaction]);
        }
        const transactionExplanation = (await this.explainTransaction({
            txHex: signedTransaction,
        }));
        transactionExplanation.txHex = signedTransaction;
        if (isKrsRecovery) {
            transactionExplanation.backupKey = params.backupKey;
            transactionExplanation.coin = this.getChain();
        }
        return transactionExplanation;
    }
    /**
     * Generate a new keypair for this coin.
     * @param seed Seed from which the new keypair should be generated, otherwise a random seed is used
     */
    generateKeyPair(seed) {
        const keyPair = seed ? new keyPair_1.KeyPair({ seed }) : new keyPair_1.KeyPair();
        const keys = keyPair.getExtendedKeys();
        if (!keys.xprv) {
            throw new Error('Missing prv in key generation.');
        }
        return {
            pub: keys.xpub,
            prv: keys.xprv,
        };
    }
    async parseTransaction(params) {
        return {};
    }
    /** @inheritDoc */
    auditDecryptedKey(params) {
        throw new sdk_core_1.MethodNotImplementedError();
    }
}
exports.Xrp = Xrp;
