// @bitgo/sdk-coin-xrp/dist/src/xrpToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrpToken = void 0;
const statics_1 = require("@bitgo/statics");
const xrp_1 = require("./xrp");
class XrpToken extends xrp_1.Xrp {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('xrp') : statics_1.coins.get('txrp');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new XrpToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.xrp.tokens, ...statics_1.tokens.testnet.xrp.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            const tokenConstructor = XrpToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get issuerAddress() {
        return this.tokenConfig.issuerAddress;
    }
    get currencyCode() {
        return this.tokenConfig.currencyCode;
    }
    get contractAddress() {
        return this.tokenConfig.contractAddress;
    }
    get domain() {
        return this.tokenConfig.domain || '';
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Xrp Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.XrpToken = XrpToken;
