// @bitgo/sdk-coin-algo/dist/test/integration/algo.integration.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const should = __importStar(require("should"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
describe('ALGO:', function () {
    let bitgo;
    let algocoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
        algocoin = bitgo.coin('talgo');
    });
    describe('Algo non partipation key reg transaction', function () {
        // TODO: The test currently works against 'testnet-01' environment as the accompanying platform changes are not
        // yet available in 'test' environment.
        // Once the platform updates are available in 'test', this test will be enabled, updated and a separate PR will be submitted
        xit('should successfully submit non paticipating key reg transaction', async function () {
            const algoWallet = await algocoin.wallets().getWallet({ id: sdk_test_1.TestBitGo.V2.TEST_ALGO_WALLET_ID });
            // Build and sign the transaction
            const preBuiltSignedTx = await algoWallet.prebuildAndSignTransaction({
                type: 'keyreg',
                nonParticipation: true,
                walletPassphrase: sdk_test_1.TestBitGo.TEST_PASSWORD,
            });
            preBuiltSignedTx.should.have.property('txHex');
            preBuiltSignedTx.should.have.propertyByPath('txInfo', 'type').eql('keyreg');
            preBuiltSignedTx.should.have.propertyByPath('txInfo', 'nonParticipation').eql(true);
            // submit the transaction
            const txResponse = await algoWallet.submitTransaction({ halfSigned: preBuiltSignedTx.halfSigned });
            should.exist(txResponse);
            txResponse.should.have.property('transfer', 'id');
            txResponse.should.have.property('txid');
            txResponse.should.have.property('tx');
            txResponse.should.have.property('status').eql('signed');
        });
    });
});
