// @bitgo/abstract-lightning/dist/src/wallet/lightning.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightningWallet = void 0;
exports.getLightningKeychain = getLightningKeychain;
exports.getLightningAuthKeychains = getLightningAuthKeychains;
const sdkcore = __importStar(require("@bitgo/sdk-core"));
const sdk_core_1 = require("@bitgo/sdk-core");
const t = __importStar(require("io-ts"));
const lightning_1 = require("../lightning");
const codecs_1 = require("../codecs");
const public_types_1 = require("@bitgo/public-types");
/**
 * Get the lightning keychain for the given wallet.
 */
async function getLightningKeychain(wallet) {
    const coin = wallet.baseCoin;
    if (coin.getFamily() !== 'lnbtc') {
        throw new Error(`Invalid coin to get lightning wallet key: ${coin.getFamily()}`);
    }
    const keyIds = wallet.keyIds();
    if (keyIds.length !== 1) {
        throw new Error(`Invalid number of key in lightning wallet: ${keyIds.length}`);
    }
    const keychain = await coin.keychains().get({ id: keyIds[0] });
    return sdkcore.decodeOrElse(codecs_1.LightningKeychain.name, codecs_1.LightningKeychain, keychain, (_) => {
        throw new Error(`Invalid user key`);
    });
}
/**
 * Get the lightning auth keychains for the given wallet.
 */
async function getLightningAuthKeychains(wallet) {
    const coin = wallet.baseCoin;
    if (coin.getFamily() !== 'lnbtc') {
        throw new Error(`Invalid coin to get lightning wallet auth keys: ${coin.getFamily()}`);
    }
    const authKeyIds = wallet.coinSpecific()?.keys;
    if (authKeyIds?.length !== 2) {
        throw new Error(`Invalid number of auth keys in lightning wallet: ${authKeyIds?.length}`);
    }
    const keychains = await Promise.all(authKeyIds.map((id) => coin.keychains().get({ id })));
    const authKeychains = keychains.map((keychain) => {
        return sdkcore.decodeOrElse(codecs_1.LightningAuthKeychain.name, codecs_1.LightningAuthKeychain, keychain, (_) => {
            // DON'T throw errors from decodeOrElse. It could leak sensitive information.
            throw new Error(`Invalid lightning auth key: ${keychain?.id}`);
        });
    });
    const [userAuthKey, nodeAuthKey] = ['userAuth', 'nodeAuth'].map((purpose) => {
        const keychain = authKeychains.find((k) => (0, lightning_1.unwrapLightningCoinSpecific)(k.coinSpecific, coin.getChain()).purpose === purpose);
        if (!keychain) {
            throw new Error(`Missing ${purpose} key`);
        }
        return keychain;
    });
    return { userAuthKey, nodeAuthKey };
}
class LightningWallet {
    constructor(wallet) {
        const coin = wallet.baseCoin;
        if (coin.getFamily() !== 'lnbtc') {
            throw new Error(`Invalid coin for lightning wallet: ${coin.getFamily()}`);
        }
        this.wallet = wallet;
    }
    async createInvoice(params) {
        const createInvoiceResponse = await this.wallet.bitgo
            .post(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/invoice`, 2))
            .send(t.exact(codecs_1.CreateInvoiceBody).encode(params))
            .result();
        return sdkcore.decodeOrElse(codecs_1.Invoice.name, codecs_1.Invoice, createInvoiceResponse, (error) => {
            // DON'T throw errors from decodeOrElse. It could leak sensitive information.
            throw new Error(`Invalid create invoice response ${error}`);
        });
    }
    async getInvoice(paymentHash) {
        const response = await this.wallet.bitgo
            .get(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/invoice/${paymentHash}`, 2))
            .result();
        return (0, sdk_core_1.decodeOrElse)(codecs_1.Invoice.name, codecs_1.Invoice, response, (error) => {
            throw new Error(`Invalid get invoice response ${error}`);
        });
    }
    async listInvoices(params) {
        const returnCodec = codecs_1.ListInvoicesResponse;
        const createInvoiceResponse = await this.wallet.bitgo
            .get(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/invoice`, 2))
            .query(codecs_1.InvoiceQuery.encode(params))
            .result();
        return sdkcore.decodeOrElse(returnCodec.name, returnCodec, createInvoiceResponse, (error) => {
            throw new Error(`Invalid list invoices response ${error}`);
        });
    }
    async payInvoice(params) {
        const reqId = new sdk_core_1.RequestTracer();
        this.wallet.bitgo.setRequestTracer(reqId);
        const { userAuthKey } = await getLightningAuthKeychains(this.wallet);
        const userAuthKeyEncryptedPrv = userAuthKey.encryptedPrv;
        if (!userAuthKeyEncryptedPrv) {
            throw new Error(`user auth key is missing encrypted private key`);
        }
        const signature = (0, lightning_1.createMessageSignature)(t.exact(public_types_1.LightningPaymentRequest).encode(params), this.wallet.bitgo.decrypt({ password: params.passphrase, input: userAuthKeyEncryptedPrv }));
        const paymentIntent = {
            intent: {
                comment: params.comment,
                sequenceId: params.sequenceId,
                intentType: 'payment',
                signedRequest: {
                    invoice: params.invoice,
                    amountMsat: params.amountMsat,
                    feeLimitMsat: params.feeLimitMsat,
                    feeLimitRatio: params.feeLimitRatio,
                },
                signature,
            },
        };
        const transactionRequestCreate = (await this.wallet.bitgo
            .post(this.wallet.bitgo.url('/wallet/' + this.wallet.id() + '/txrequests', 2))
            .send(t.type({ intent: public_types_1.LightningPaymentIntent }).encode(paymentIntent))
            .result());
        if (transactionRequestCreate.state === 'pendingApproval') {
            const pendingApprovals = new sdk_core_1.PendingApprovals(this.wallet.bitgo, this.wallet.baseCoin);
            const pendingApproval = await pendingApprovals.get({ id: transactionRequestCreate.pendingApprovalId });
            return {
                pendingApproval: pendingApproval.toJSON(),
                txRequestId: transactionRequestCreate.txRequestId,
                txRequestState: transactionRequestCreate.state,
            };
        }
        await this.wallet.bitgo
            .post(this.wallet.bitgo.url('/wallet/' + this.wallet.id() + '/txrequests/' + transactionRequestCreate.txRequestId + '/transfers', 2))
            .send()
            .result();
        const transactionRequestSend = await sdk_core_1.commonTssMethods.sendTxRequest(this.wallet.bitgo, this.wallet.id(), transactionRequestCreate.txRequestId, sdk_core_1.RequestType.tx, reqId);
        const coinSpecific = transactionRequestSend.transactions?.[0]?.unsignedTx?.coinSpecific;
        return {
            txRequestId: transactionRequestCreate.txRequestId,
            txRequestState: transactionRequestSend.state,
            paymentStatus: coinSpecific
                ? t.exact(codecs_1.LndCreatePaymentResponse).encode(coinSpecific)
                : undefined,
        };
    }
    async withdrawOnchain(params) {
        const reqId = new sdk_core_1.RequestTracer();
        this.wallet.bitgo.setRequestTracer(reqId);
        const paymentIntent = {
            intent: {
                comment: params.comment,
                sequenceId: params.sequenceId,
                onchainRequest: {
                    recipients: params.recipients,
                    satsPerVbyte: params.satsPerVbyte,
                    numBlocks: params.numBlocks,
                },
                intentType: 'payment',
            },
        };
        const transactionRequestCreate = (await this.wallet.bitgo
            .post(this.wallet.bitgo.url('/wallet/' + this.wallet.id() + '/txrequests', 2))
            .send(t.type({ intent: public_types_1.LightningPaymentIntent }).encode(paymentIntent))
            .result());
        if (!transactionRequestCreate.transactions ||
            transactionRequestCreate.transactions.length === 0 ||
            !transactionRequestCreate.transactions[0].unsignedTx.serializedTxHex) {
            throw new Error(`serialized txHex is missing`);
        }
        const walletData = this.wallet.toJSON();
        if (!walletData.coinSpecific.watchOnlyAccounts) {
            throw new Error(`wallet is missing watch only accounts`);
        }
        const watchOnlyAccountDetails = (0, sdk_core_1.decodeOrElse)(codecs_1.WatchOnly.name, codecs_1.WatchOnly, walletData.coinSpecific.watchOnlyAccounts, (errors) => {
            throw new Error(`invalid watch only accounts, error: ${errors}`);
        });
        const network = (0, lightning_1.getUtxolibNetwork)(this.wallet.coin());
        try {
            (0, lightning_1.validatePsbtForWithdraw)(transactionRequestCreate.transactions[0].unsignedTx.serializedTxHex, network, params.recipients, watchOnlyAccountDetails.accounts);
        }
        catch (err) {
            throw new Error(`error validating withdraw psbt: ${err}`);
        }
        const { userAuthKey } = await getLightningAuthKeychains(this.wallet);
        const userAuthKeyEncryptedPrv = userAuthKey.encryptedPrv;
        if (!userAuthKeyEncryptedPrv) {
            throw new Error(`user auth key is missing encrypted private key`);
        }
        const signature = (0, lightning_1.createMessageSignature)(transactionRequestCreate.transactions[0].unsignedTx.serializedTxHex, this.wallet.bitgo.decrypt({ password: params.passphrase, input: userAuthKeyEncryptedPrv }));
        const transactionRequestWithSignature = (await this.wallet.bitgo
            .put(this.wallet.bitgo.url('/wallet/' + this.wallet.id() + '/txrequests/' + transactionRequestCreate.txRequestId + '/coinSpecific', 2))
            .send({
            unsignedCoinSpecific: {
                coin: this.wallet.baseCoin.getChain(),
                data: { signature },
            },
        })
            .result());
        if (transactionRequestWithSignature.state === 'pendingApproval') {
            const pendingApprovals = new sdk_core_1.PendingApprovals(this.wallet.bitgo, this.wallet.baseCoin);
            const pendingApproval = await pendingApprovals.get({ id: transactionRequestWithSignature.pendingApprovalId });
            return {
                pendingApproval: pendingApproval.toJSON(),
                txRequestId: transactionRequestWithSignature.txRequestId,
                txRequestState: transactionRequestWithSignature.state,
            };
        }
        await this.wallet.bitgo
            .post(this.wallet.bitgo.url('/wallet/' + this.wallet.id() + '/txrequests/' + transactionRequestWithSignature.txRequestId + '/transfers', 2))
            .send()
            .result();
        const transactionRequestSend = await sdk_core_1.commonTssMethods.sendTxRequest(this.wallet.bitgo, this.wallet.id(), transactionRequestWithSignature.txRequestId, sdk_core_1.RequestType.tx, reqId);
        const coinSpecific = transactionRequestSend.transactions?.[0]?.unsignedTx?.coinSpecific;
        return {
            txRequestId: transactionRequestWithSignature.txRequestId,
            txRequestState: transactionRequestSend.state,
            withdrawStatus: coinSpecific && 'status' in coinSpecific
                ? t.exact(codecs_1.LndCreateWithdrawResponse).encode(coinSpecific)
                : undefined,
        };
    }
    async getPayment(paymentId) {
        const response = await this.wallet.bitgo
            .get(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/payment/${paymentId}`, 2))
            .result();
        return (0, sdk_core_1.decodeOrElse)(codecs_1.PaymentInfo.name, codecs_1.PaymentInfo, response, (error) => {
            throw new Error(`Invalid payment response: ${error}`);
        });
    }
    async listPayments(params) {
        const response = await this.wallet.bitgo
            .get(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/payment`, 2))
            .query(codecs_1.PaymentQuery.encode(params))
            .result();
        return (0, sdk_core_1.decodeOrElse)(codecs_1.ListPaymentsResponse.name, codecs_1.ListPaymentsResponse, response, (error) => {
            throw new Error(`Invalid payment list response: ${error}`);
        });
    }
    async getTransaction(txId) {
        const response = await this.wallet.bitgo
            .get(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/transaction/${txId}`, 2))
            .result();
        return (0, sdk_core_1.decodeOrElse)(codecs_1.Transaction.name, codecs_1.Transaction, response, (error) => {
            throw new Error(`Invalid transaction response: ${error}`);
        });
    }
    async listTransactions(params) {
        const response = await this.wallet.bitgo
            .get(this.wallet.bitgo.url(`/wallet/${this.wallet.id()}/lightning/transaction`, 2))
            .query(codecs_1.TransactionQuery.encode(params))
            .result();
        return (0, sdk_core_1.decodeOrElse)(codecs_1.ListTransactionsResponse.name, codecs_1.ListTransactionsResponse, response, (error) => {
            throw new Error(`Invalid transaction list response: ${error}`);
        });
    }
}
exports.LightningWallet = LightningWallet;
