// @bitgo/abstract-lightning/dist/src/wallet/selfCustodialLightning.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfCustodialLightningWallet = void 0;
exports.updateWalletCoinSpecific = updateWalletCoinSpecific;
const sdkcore = __importStar(require("@bitgo/sdk-core"));
const codecs_1 = require("../codecs");
const lightning_1 = require("./lightning");
const lightning_2 = require("../lightning");
const t = __importStar(require("io-ts"));
function encryptWalletUpdateRequest(wallet, params, userAuthKeyEncryptedPrv) {
    const coinName = wallet.coin();
    const requestWithEncryption = {
        ...params,
    };
    const userAuthXprv = wallet.bitgo.decrypt({
        password: params.passphrase,
        input: userAuthKeyEncryptedPrv,
    });
    if (params.signerTlsKey) {
        requestWithEncryption.encryptedSignerTlsKey = wallet.bitgo.encrypt({
            password: params.passphrase,
            input: params.signerTlsKey,
        });
    }
    if (params.signerAdminMacaroon) {
        requestWithEncryption.encryptedSignerAdminMacaroon = wallet.bitgo.encrypt({
            password: params.passphrase,
            input: params.signerAdminMacaroon,
        });
    }
    if (params.signerMacaroon) {
        requestWithEncryption.encryptedSignerMacaroon = wallet.bitgo.encrypt({
            password: (0, lightning_2.deriveLightningServiceSharedSecret)(coinName, userAuthXprv).toString('hex'),
            input: params.signerMacaroon,
        });
    }
    return t.exact(codecs_1.UpdateLightningWalletEncryptedRequest).encode(requestWithEncryption);
}
/**
 * Updates the coin-specific configuration for a Lightning Wallet.
 *
 * @param {Wallet} wallet - Wallet.
 * @param {UpdateLightningWalletClientRequest} params - The parameters containing the updated wallet-specific details.
 *   - `encryptedSignerMacaroon` (optional): This macaroon is used by the watch-only node to ask the signer node to sign transactions.
 *     Encrypted with ECDH secret key from private key of wallet's user auth key and public key of lightning service.
 *   - `encryptedSignerAdminMacaroon` (optional): Generated when initializing the wallet of the signer node.
 *     Encrypted with client's wallet passphrase.
 *   - `signerHost` (optional): The host address of the Lightning signer node.
 *   - `encryptedSignerTlsKey` (optional): The wallet passphrase encrypted TLS key of the signer.
 *   - `passphrase` (required): The wallet passphrase.
 *   - `signerTlsCert` (optional): The TLS certificate of the signer.
 *   - `watchOnlyAccounts` (optional): These are the accounts used to initialize the watch-only wallet.
 * @returns {Promise<unknown>} A promise resolving to the updated wallet response or throwing an error if the update fails.
 */
async function updateWalletCoinSpecific(wallet, params) {
    if (!(0, lightning_2.isLightningCoinName)(wallet.coin())) {
        throw new Error(`cant update lightning wallet coin specific for coin ${wallet.subType()}`);
    }
    if (wallet.subType() !== 'lightningSelfCustody') {
        throw new Error(`cant update lightning wallet coin specific for wallet type ${wallet.subType()}`);
    }
    sdkcore.decodeOrElse(codecs_1.UpdateLightningWalletClientRequest.name, codecs_1.UpdateLightningWalletClientRequest, params, (errors) => {
        // DON'T throw errors from decodeOrElse. It could leak sensitive information.
        throw new Error(`Invalid params for lightning specific update wallet`);
    });
    const { userAuthKey } = await (0, lightning_1.getLightningAuthKeychains)(wallet);
    const userAuthKeyEncryptedPrv = userAuthKey.encryptedPrv;
    if (!userAuthKeyEncryptedPrv) {
        throw new Error(`user auth key is missing encrypted private key`);
    }
    const updateRequestWithEncryption = encryptWalletUpdateRequest(wallet, params, userAuthKeyEncryptedPrv);
    const signature = (0, lightning_2.createMessageSignature)(updateRequestWithEncryption, wallet.bitgo.decrypt({ password: params.passphrase, input: userAuthKeyEncryptedPrv }));
    const coinSpecific = {
        [wallet.coin()]: {
            signedRequest: updateRequestWithEncryption,
            signature,
        },
    };
    return await wallet.bitgo.put(wallet.url()).send({ coinSpecific }).result();
}
class SelfCustodialLightningWallet extends lightning_1.LightningWallet {
    constructor(wallet) {
        super(wallet);
        if (wallet.subType() !== 'lightningSelfCustody') {
            throw new Error(`Invalid lightning wallet type for self custodial lightning: ${wallet.subType()}`);
        }
    }
    async getChannelBackup() {
        const backupResponse = await this.wallet.bitgo
            .get(this.wallet.baseCoin.url(`/wallet/${this.wallet.id()}/lightning/backup`))
            .result();
        return sdkcore.decodeOrElse(codecs_1.BackupResponse.name, codecs_1.BackupResponse, backupResponse, (error) => {
            throw new Error(`Invalid backup response: ${error}`);
        });
    }
}
exports.SelfCustodialLightningWallet = SelfCustodialLightningWallet;
