// @bitgo/abstract-utxo/dist/cjs/src/transaction/descriptor/explainPsbt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.explainPsbt = explainPsbt;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const wasmUtil_1 = require("../../wasmUtil");
function toRecipient(output, coinName) {
    const address = output.address ?? `scriptPubKey:${Buffer.from(output.script).toString('hex')}`;
    return {
        address,
        amount: output.value.toString(),
    };
}
function getInputSignaturesForInputIndex(psbt, inputIndex) {
    if (!psbt.hasPartialSignatures(inputIndex)) {
        return 0;
    }
    const partialSigs = psbt.getPartialSignatures(inputIndex);
    return partialSigs.reduce((agg, p) => {
        const valid = psbt.validateSignatureAtInput(inputIndex, p.pubkey);
        return agg + (valid ? 1 : 0);
    }, 0);
}
function getInputSignatures(psbt) {
    return Array.from({ length: psbt.inputCount() }, (_, i) => getInputSignaturesForInputIndex(psbt, i));
}
function explainPsbt(psbt, descriptors, coinName) {
    const parsedTransaction = wasm_utxo_1.descriptorWallet.parse(psbt, descriptors, coinName);
    const { inputs, outputs } = parsedTransaction;
    const externalOutputs = outputs.filter((o) => o.scriptId === undefined);
    const changeOutputs = outputs.filter((o) => o.scriptId !== undefined);
    const fee = (0, wasmUtil_1.sumValues)(inputs) - (0, wasmUtil_1.sumValues)(outputs);
    const inputSignatures = getInputSignatures(psbt);
    return {
        inputSignatures,
        signatures: inputSignatures.reduce((a, b) => Math.min(a, b), Infinity),
        locktime: psbt.lockTime(),
        id: psbt.unsignedTxId(),
        outputs: externalOutputs.map((o) => toRecipient(o, coinName)),
        outputAmount: (0, wasmUtil_1.sumValues)(externalOutputs).toString(),
        changeOutputs: changeOutputs.map((o) => toRecipient(o, coinName)),
        changeAmount: (0, wasmUtil_1.sumValues)(changeOutputs).toString(),
        fee: fee.toString(),
    };
}
