// @bitgo/abstract-utxo/dist/cjs/src/transaction/descriptor/verifyTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateValidationError = exports.ErrorImplicitExternalOutputs = exports.ErrorMissingOutputs = exports.ValidationError = void 0;
exports.assertExpectedOutputDifference = assertExpectedOutputDifference;
exports.assertValidTransaction = assertValidTransaction;
exports.verifyTransaction = verifyTransaction;
const sdk_core_1 = require("@bitgo/sdk-core");
const wasmUtil_1 = require("../../wasmUtil");
const parse_1 = require("./parse");
class ValidationError extends Error {
    constructor(message) {
        super(message);
    }
}
exports.ValidationError = ValidationError;
class ErrorMissingOutputs extends ValidationError {
    constructor(missingOutputs) {
        super(`missing outputs (count=${missingOutputs.length})`);
        this.missingOutputs = missingOutputs;
    }
}
exports.ErrorMissingOutputs = ErrorMissingOutputs;
class ErrorImplicitExternalOutputs extends ValidationError {
    constructor(implicitExternalOutputs) {
        super(`unexpected implicit external outputs (count=${implicitExternalOutputs.length})`);
        this.implicitExternalOutputs = implicitExternalOutputs;
    }
}
exports.ErrorImplicitExternalOutputs = ErrorImplicitExternalOutputs;
class AggregateValidationError extends ValidationError {
    constructor(errors) {
        super(`aggregate validation error (count=${errors.length})`);
        this.errors = errors;
    }
}
exports.AggregateValidationError = AggregateValidationError;
function assertExpectedOutputDifference(parsedOutputs) {
    const errors = [];
    if (parsedOutputs.missingOutputs.length > 0) {
        errors.push(new ErrorMissingOutputs(parsedOutputs.missingOutputs));
    }
    if (parsedOutputs.implicitExternalOutputs.length > 0) {
        // FIXME: for paygo we need to relax this a little bit
        errors.push(new ErrorImplicitExternalOutputs(parsedOutputs.implicitExternalOutputs));
    }
    if (errors.length > 0) {
        // FIXME(BTC-1688): enable ES2021
        // throw new AggregateError(errors);
        throw new AggregateValidationError(errors);
    }
}
function assertValidTransaction(psbt, descriptors, recipients, coinName) {
    assertExpectedOutputDifference((0, parse_1.toBaseParsedTransactionOutputsFromPsbt)(psbt, descriptors, recipients, coinName));
}
/**
 * Wrapper around assertValidTransaction that returns a boolean instead of throwing.
 *
 * We follow the AbstractUtxoCoin interface here which is a bit confused - the return value is a boolean but we
 * also throw errors (because we actually want to know what went wrong).
 *
 * @param coin
 * @param params
 * @param descriptorMap
 * @returns {boolean} True if verification passes
 * @throws {TxIntentMismatchError} if transaction validation fails
 */
async function verifyTransaction(coin, params, descriptorMap) {
    const tx = coin.decodeTransactionFromPrebuild(params.txPrebuild);
    let psbt;
    try {
        psbt = (0, wasmUtil_1.toWasmPsbt)(tx);
    }
    catch (e) {
        const txExplanation = await sdk_core_1.TxIntentMismatchError.tryGetTxExplanation(coin, params.txPrebuild);
        const errorDetail = e instanceof Error ? e.message : String(e);
        throw new sdk_core_1.TxIntentMismatchError(`unexpected transaction type: ${errorDetail}`, params.reqId, [params.txParams], params.txPrebuild.txHex, txExplanation);
    }
    assertValidTransaction(psbt, descriptorMap, params.txParams.recipients ?? [], coin.name);
    return true;
}
