// @bitgo/sdk-core/dist/src/bitgo/address-book/address-book.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressBook = void 0;
class AddressBook {
    constructor(enterpriseId, bitgo, wallet) {
        this.enterpriseId = enterpriseId;
        this.wallet = wallet;
        this.bitgo = bitgo;
    }
    listing() {
        /**
         * TODO(PX-2794): Move to structuredClone
         * https://github.com/BitGo/BitGoJS/pull/4119
         */
        return JSON.parse(JSON.stringify(this._listing));
    }
    /**
     * Get a list of connections the wallet has made to other directory or manually added contacts.
     */
    getConnections(params) {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/connections');
        return this.bitgo.get(url).set('enterprise-id', this.enterpriseId).send(params).result();
    }
    /**
     * Create a connection between an enterprise listing entry (wallet) to another listing entry
     * @param params
     * @param {string} listingEntryId Your enterprise listing entry id. Requires the creation of a listing entry before use.
     * @param {string=} localListingEntryDescription Optional name to override the name of the counterparties listing entry.
     * @param {string} targetListingEntryId If you know the other parties listing entry id
     * @param {string} walletId If you don't know the targetListingEntryId and are adding manually
     * @param {string} localListingName Required if using walletId
     * @returns {Promise<CreateAddressBookConnectionResponse>}
     */
    createConnection(params) {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/connections');
        return this.bitgo.post(url).set('enterprise-id', this.enterpriseId).send(params).result();
    }
    /**
     * Update one or many connections to a new status
     */
    updateConnection(params) {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/connections');
        return this.bitgo.put(url).set('enterprise-id', this.enterpriseId).send(params).result();
    }
    /**
     * Get the address book listing for the enterprise
     */
    async getListing() {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/listing/global');
        const response = await this.bitgo
            .get(url)
            .set('enterprise-id', this.enterpriseId)
            .send()
            .result();
        this._listing = response;
        return this.listing();
    }
    /**
     * Create the listing used for each wallet's listing entry
     */
    async createListing(params) {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/listing/global');
        const response = await this.bitgo
            .post(url)
            .set('enterprise-id', this.enterpriseId)
            .send(params)
            .result();
        this._listing = { ...response, listingEntries: [] };
        return this.listing();
    }
    /**
     * Update the name and description of the listing
     */
    async updateListing({ listingId, ...params }) {
        const url = this.bitgo.microservicesUrl(`/api/address-book/v1/listing/${listingId}`);
        const response = await this.bitgo
            .put(url)
            .set('enterprise-id', this.enterpriseId)
            .send(params)
            .result();
        this._listing = { ...response, listingEntries: this._listing?.listingEntries ? this._listing.listingEntries : [] };
        return this.listing();
    }
    /**
     * Return a list of listing entry contacts that are connected to your enterprise listing entries (wallets)
     */
    getListingEntryContacts(params) {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/listing/entry/contacts');
        return this.bitgo.get(url).set('enterprise-id', this.enterpriseId).send(params).result();
    }
    /**
     * Return a public list of other listing entries that you can connect with.
     */
    getListingEntryDirectory(params) {
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/listing/entry/directory');
        return this.bitgo.get(url).set('enterprise-id', this.enterpriseId).send(params).result();
    }
    /**
     * Create a listing entry for use in the public directory or keep private and share the listing entry id with others.
     */
    async createListingEntry(params) {
        if (!params.walletId && this.wallet) {
            params.walletId = this.wallet.id();
        }
        const url = this.bitgo.microservicesUrl('/api/address-book/v1/listing/entry/global');
        const response = await this.bitgo
            .post(url)
            .set('enterprise-id', this.enterpriseId)
            .send(params)
            .result();
        if (this._listing) {
            this._listing.listingEntries?.push({ ...response });
        }
        return response;
    }
    /**
     * Update a listing entry (wallet)
     */
    async updateListingEntry({ listingEntryId, ...params }) {
        const url = this.bitgo.microservicesUrl(`/api/address-book/v1/listing/entry/${listingEntryId}`);
        const response = await this.bitgo
            .put(url)
            .set('enterprise-id', this.enterpriseId)
            .send(params)
            .result();
        if (this._listing) {
            const index = this._listing?.listingEntries?.findIndex((x) => x.id === response.id) ?? -1;
            if (index > -1 && this._listing.listingEntries) {
                this._listing.listingEntries[index] = { ...response };
            }
            else {
                this._listing.listingEntries?.push({ ...response });
            }
        }
        return response;
    }
}
exports.AddressBook = AddressBook;
