// @bitgo/sdk-coin-hash/dist/test/unit/hash.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const sdk_test_1 = require("@bitgo/sdk-test");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const mocha_1 = require("mocha");
const sinon_1 = __importDefault(require("sinon"));
const src_1 = require("../../src");
const utils_1 = require("../../src/lib/utils");
const hash_1 = require("../resources/hash");
const should = require("should");
describe('HASH', function () {
    let bitgo;
    let hash;
    let thash;
    let mainnetUtils;
    let testnetUtils;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('hash', src_1.Hash.createInstance);
        bitgo.safeRegister('thash', src_1.Thash.createInstance);
        bitgo.initializeTestVars();
        hash = bitgo.coin('hash');
        thash = bitgo.coin('thash');
        mainnetUtils = new utils_1.HashUtils(statics_1.NetworkType.MAINNET);
        testnetUtils = new utils_1.HashUtils(statics_1.NetworkType.TESTNET);
    });
    it('should return the right info', function () {
        const hash = bitgo.coin('hash');
        const thash = bitgo.coin('thash');
        hash.getChain().should.equal('hash');
        hash.getFamily().should.equal('hash');
        hash.getFullName().should.equal('Provenance');
        hash.getBaseFactor().should.equal(1e9);
        thash.getChain().should.equal('thash');
        thash.getFamily().should.equal('hash');
        thash.getFullName().should.equal('Testnet Provenance');
        thash.getBaseFactor().should.equal(1e9);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const mainnetAddressDetails = hash.getAddressDetails(hash_1.mainnetAddress.noMemoIdAddress);
            mainnetAddressDetails.address.should.equal(hash_1.mainnetAddress.noMemoIdAddress);
            should.not.exist(mainnetAddressDetails.memoId);
            const testnetAddressDetails = thash.getAddressDetails(hash_1.testnetAddress.noMemoIdAddress);
            testnetAddressDetails.address.should.equal(hash_1.testnetAddress.noMemoIdAddress);
            should.not.exist(testnetAddressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const mainnetAddressDetails = hash.getAddressDetails(hash_1.mainnetAddress.validMemoIdAddress);
            mainnetAddressDetails.address.should.equal(hash_1.mainnetAddress.validMemoIdAddress.split('?')[0]);
            mainnetAddressDetails.memoId.should.equal('2');
            const testnetAddressDetails = thash.getAddressDetails(hash_1.testnetAddress.validMemoIdAddress);
            testnetAddressDetails.address.should.equal(hash_1.testnetAddress.validMemoIdAddress.split('?')[0]);
            testnetAddressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                hash.getAddressDetails(hash_1.mainnetAddress.multipleMemoIdAddress);
            }).should.throw();
            (() => {
                thash.getAddressDetails(hash_1.testnetAddress.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const receiveAddress = {
                address: 'tp1496r8u4a48k6khknrhzd6c8cm3c64ewxy5p2rj?memoId=7',
                coinSpecific: {
                    rootAddress: 'tp1496r8u4a48k6khknrhzd6c8cm3c64ewxy5p2rj',
                    memoID: '7',
                },
            };
            const isValid = await thash.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.address1), true);
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.address2), true);
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.address3), false);
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.address4), false);
            should.equal(mainnetUtils.isValidAddress('dfjk35y'), false);
            should.equal(mainnetUtils.isValidAddress(undefined), false);
            should.equal(mainnetUtils.isValidAddress(''), false);
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.validMemoIdAddress), true);
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.multipleMemoIdAddress), false);
            should.equal(mainnetUtils.isValidAddress(hash_1.mainnetAddress.groupPolicyAddress), true);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.address1), true);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.address2), true);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.address3), false);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.address4), false);
            should.equal(testnetUtils.isValidAddress('dfjk35y'), false);
            should.equal(testnetUtils.isValidAddress(undefined), false);
            should.equal(testnetUtils.isValidAddress(''), false);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.validMemoIdAddress), true);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.multipleMemoIdAddress), false);
            should.equal(testnetUtils.isValidAddress(hash_1.testnetAddress.groupPolicyAddress), true);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(mainnetUtils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress1), true);
            should.equal(mainnetUtils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress2), true);
            should.equal(mainnetUtils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress3), false);
            should.equal(mainnetUtils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress4), false);
            should.equal(mainnetUtils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(mainnetUtils.isValidValidatorAddress(undefined), false);
            should.equal(mainnetUtils.isValidValidatorAddress(''), false);
            should.equal(testnetUtils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress1), true);
            should.equal(testnetUtils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress2), true);
            should.equal(testnetUtils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress3), false);
            should.equal(testnetUtils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress4), false);
            should.equal(testnetUtils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(testnetUtils.isValidValidatorAddress(undefined), false);
            should.equal(testnetUtils.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: hash_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: hash_1.TEST_SEND_TX.recipient,
                        amount: hash_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await thash.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify delegate transaction', async function () {
            const txPrebuild = {
                txHex: hash_1.TEST_DELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: hash_1.TEST_DELEGATE_TX.validator,
                        amount: hash_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await thash.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify undelegate transaction', async function () {
            const txPrebuild = {
                txHex: hash_1.TEST_UNDELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: hash_1.TEST_UNDELEGATE_TX.validator,
                        amount: hash_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await thash.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify withdraw rewards transaction', async function () {
            const txPrebuild = {
                txHex: hash_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: hash_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await thash.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await thash
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await thash.explainTransaction({
                txHex: hash_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: hash_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: hash_1.TEST_SEND_TX.recipient,
                        amount: hash_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: hash_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: hash_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a delegate transaction', async function () {
            const explainedTransaction = await thash.explainTransaction({
                txHex: hash_1.TEST_DELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: hash_1.TEST_DELEGATE_TX.hash,
                outputs: [
                    {
                        address: hash_1.TEST_DELEGATE_TX.validator,
                        amount: hash_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: hash_1.TEST_DELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: hash_1.TEST_DELEGATE_TX.gasBudget.amount[0].amount },
                type: 13,
            });
        });
        it('should explain a undelegate transaction', async function () {
            const explainedTransaction = await thash.explainTransaction({
                txHex: hash_1.TEST_UNDELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: hash_1.TEST_UNDELEGATE_TX.hash,
                outputs: [
                    {
                        address: hash_1.TEST_UNDELEGATE_TX.validator,
                        amount: hash_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: hash_1.TEST_UNDELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: hash_1.TEST_UNDELEGATE_TX.gasBudget.amount[0].amount },
                type: 17,
            });
        });
        it('should explain a withdraw transaction', async function () {
            const explainedTransaction = await thash.explainTransaction({
                txHex: hash_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: hash_1.TEST_WITHDRAW_REWARDS_TX.hash,
                outputs: [
                    {
                        address: hash_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
                outputAmount: undefined,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: hash_1.TEST_WITHDRAW_REWARDS_TX.gasBudget.amount[0].amount },
                type: 15,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await thash.explainTransaction({
                txHex: hash_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: hash_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: hash_1.TEST_TX_WITH_MEMO.to,
                        amount: hash_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: hash_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: hash_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: hash_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await thash.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await thash.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        it('should parse a transfer transaction', async function () {
            const transferInputsResponse = {
                address: hash_1.TEST_SEND_TX.recipient,
                amount: new bignumber_js_1.default(hash_1.TEST_SEND_TX.sendAmount).plus(hash_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
            };
            const transferOutputsResponse = {
                address: hash_1.TEST_SEND_TX.recipient,
                amount: hash_1.TEST_SEND_TX.sendAmount,
            };
            const parsedTransaction = await thash.parseTransaction({ txHex: hash_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Hash.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await thash
                .parseTransaction({ txHex: hash_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
    describe('Recover transaction: success path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = hash_1.wrwUser.destinationAddress;
        const coin = statics_1.coins.get('thash');
        const testBalance = [
            {
                denom: 'nhash',
                amount: '150000000000',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.Hash.prototype, 'getAccountBalance');
            accountBalance.withArgs(hash_1.wrwUser.senderAddress).resolves(testBalance);
            const accountDetails = sandBox.stub(src_1.Hash.prototype, 'getAccountDetails');
            accountDetails.withArgs(hash_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Hash.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should recover funds for non-bitgo recoveries', async function () {
            const res = await thash.recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(thash.getAccountBalance);
            sandBox.assert.calledOnce(thash.getAccountDetails);
            sandBox.assert.calledOnce(thash.getChainId);
            const txn = new abstract_cosmos_1.CosmosTransaction(coin, testnetUtils);
            txn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const txnJson = txn.toJson();
            const sendMessage = txnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const gasAmount = new bignumber_js_1.default(5000000000);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.toAddress, destinationAddress);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
            should.equal(sendMessage.amount[0].denom, testBalance[0].denom);
        });
    });
    describe('Recover transaction: failure path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = hash_1.wrwUser.destinationAddress;
        const testZeroBalance = [
            {
                denom: 'nhash',
                amount: '0',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.Hash.prototype, 'getAccountBalance');
            accountBalance.withArgs(hash_1.wrwUser.senderAddress).resolves(testZeroBalance);
            const accountDetails = sandBox.stub(src_1.Hash.prototype, 'getAccountDetails');
            accountDetails.withArgs(hash_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Hash.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should throw error if backupkey is not present', async function () {
            await thash
                .recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing backupKey');
        });
        it('should throw error if userkey is not present', async function () {
            await thash
                .recover({
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing userKey');
        });
        it('should throw error if wallet passphrase is not present', async function () {
            await thash
                .recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing wallet passphrase');
        });
        it('should throw error if there is no balance', async function () {
            await thash
                .recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('Did not have enough funds to recover');
        });
    });
});
