// @bitgo/sdk-coin-hash/dist/test/unit/hashToken.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should = require("should");
const sinon_1 = __importDefault(require("sinon"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const mocha_1 = require("mocha");
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const statics_1 = require("@bitgo/statics");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const src_1 = require("../../src");
const utils_1 = require("../../src/lib/utils");
const hash_1 = require("../resources/hash");
describe('Hash Tokens', function () {
    let bitgo;
    let mainnetHashToken;
    let testnetHashToken;
    const testnetTokenName = 'thash:ylds';
    const mainnetTokenName = 'hash:ylds';
    let testnetUtils;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
        src_1.HashToken.createTokenConstructors().forEach(({ name, coinConstructor }) => {
            bitgo.safeRegister(name, coinConstructor);
        });
        bitgo.initializeTestVars();
        mainnetHashToken = bitgo.coin(mainnetTokenName);
        testnetHashToken = bitgo.coin(testnetTokenName);
        testnetUtils = new utils_1.HashUtils(statics_1.NetworkType.TESTNET);
    });
    it('should return constants for Hash YLDS testnet token', function () {
        testnetHashToken.getChain().should.equal(testnetTokenName);
        testnetHashToken.getBaseChain().should.equal('thash');
        testnetHashToken.getFullName().should.equal('Hash Token');
        testnetHashToken.getBaseFactor().should.equal(1e6);
        testnetHashToken.type.should.equal(testnetTokenName);
        testnetHashToken.name.should.equal('Testnet YLDS Token');
        testnetHashToken.coin.should.equal('thash');
        testnetHashToken.network.should.equal('Testnet');
        testnetHashToken.denom.should.equal('uylds.fcc');
        testnetHashToken.decimalPlaces.should.equal(6);
    });
    it('should return constants for Hash YLDS mainnet token', function () {
        mainnetHashToken.getChain().should.equal(mainnetTokenName);
        mainnetHashToken.getBaseChain().should.equal('hash');
        mainnetHashToken.getFullName().should.equal('Hash Token');
        mainnetHashToken.getBaseFactor().should.equal(1e6);
        mainnetHashToken.type.should.equal(mainnetTokenName);
        mainnetHashToken.name.should.equal('YLDS Token');
        mainnetHashToken.coin.should.equal('hash');
        mainnetHashToken.network.should.equal('Mainnet');
        mainnetHashToken.denom.should.equal('uylds.fcc');
        mainnetHashToken.decimalPlaces.should.equal(6);
    });
    it('should return denomination for hash tokens using hash as coinFamily', function () {
        testnetUtils.getTokenDenomsUsingCoinFamily('hash').should.deepEqual(['uylds.fcc', 'nfgrd', 'nfigr.d']);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const mainnetAddressDetails = mainnetHashToken.getAddressDetails(hash_1.mainnetAddress.noMemoIdAddress);
            mainnetAddressDetails.address.should.equal(hash_1.mainnetAddress.noMemoIdAddress);
            should.not.exist(mainnetAddressDetails.memoId);
            const testnetAddressDetails = testnetHashToken.getAddressDetails(hash_1.testnetAddress.noMemoIdAddress);
            testnetAddressDetails.address.should.equal(hash_1.testnetAddress.noMemoIdAddress);
            should.not.exist(testnetAddressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const mainnetAddressDetails = mainnetHashToken.getAddressDetails(hash_1.mainnetAddress.validMemoIdAddress);
            mainnetAddressDetails.address.should.equal(hash_1.mainnetAddress.validMemoIdAddress.split('?')[0]);
            mainnetAddressDetails.memoId.should.equal('2');
            const testnetAddressDetails = testnetHashToken.getAddressDetails(hash_1.testnetAddress.validMemoIdAddress);
            testnetAddressDetails.address.should.equal(hash_1.testnetAddress.validMemoIdAddress.split('?')[0]);
            testnetAddressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                mainnetHashToken.getAddressDetails(hash_1.mainnetAddress.multipleMemoIdAddress);
            }).should.throw();
            (() => {
                testnetHashToken.getAddressDetails(hash_1.testnetAddress.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const receiveAddress = {
                address: 'tp1496r8u4a48k6khknrhzd6c8cm3c64ewxy5p2rj?memoId=7',
                coinSpecific: {
                    rootAddress: 'tp1496r8u4a48k6khknrhzd6c8cm3c64ewxy5p2rj',
                    memoID: '7',
                },
            };
            const isValid = await testnetHashToken.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(mainnetHashToken._utils.isValidAddress(hash_1.mainnetAddress.address1), true);
            should.equal(mainnetHashToken._utils.isValidAddress(hash_1.mainnetAddress.address2), true);
            should.equal(mainnetHashToken._utils.isValidAddress(hash_1.mainnetAddress.address3), false);
            should.equal(mainnetHashToken._utils.isValidAddress(hash_1.mainnetAddress.address4), false);
            should.equal(mainnetHashToken._utils.isValidAddress('dfjk35y'), false);
            should.equal(mainnetHashToken._utils.isValidAddress(undefined), false);
            should.equal(mainnetHashToken._utils.isValidAddress(''), false);
            should.equal(mainnetHashToken._utils.isValidAddress(hash_1.mainnetAddress.validMemoIdAddress), true);
            should.equal(mainnetHashToken._utils.isValidAddress(hash_1.mainnetAddress.multipleMemoIdAddress), false);
            should.equal(testnetHashToken._utils.isValidAddress(hash_1.testnetAddress.address1), true);
            should.equal(testnetHashToken._utils.isValidAddress(hash_1.testnetAddress.address2), true);
            should.equal(testnetHashToken._utils.isValidAddress(hash_1.testnetAddress.address3), false);
            should.equal(testnetHashToken._utils.isValidAddress(hash_1.testnetAddress.address4), false);
            should.equal(testnetHashToken._utils.isValidAddress('dfjk35y'), false);
            should.equal(testnetHashToken._utils.isValidAddress(undefined), false);
            should.equal(testnetHashToken._utils.isValidAddress(''), false);
            should.equal(testnetHashToken._utils.isValidAddress(hash_1.testnetAddress.validMemoIdAddress), true);
            should.equal(testnetHashToken._utils.isValidAddress(hash_1.testnetAddress.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(mainnetHashToken._utils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress1), true);
            should.equal(mainnetHashToken._utils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress2), true);
            should.equal(mainnetHashToken._utils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress3), false);
            should.equal(mainnetHashToken._utils.isValidValidatorAddress(hash_1.mainnetAddress.validatorAddress4), false);
            should.equal(mainnetHashToken._utils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(mainnetHashToken._utils.isValidValidatorAddress(undefined), false);
            should.equal(mainnetHashToken._utils.isValidValidatorAddress(''), false);
            should.equal(testnetHashToken._utils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress1), true);
            should.equal(testnetHashToken._utils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress2), true);
            should.equal(testnetHashToken._utils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress3), false);
            should.equal(testnetHashToken._utils.isValidValidatorAddress(hash_1.testnetAddress.validatorAddress4), false);
            should.equal(testnetHashToken._utils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(testnetHashToken._utils.isValidValidatorAddress(undefined), false);
            should.equal(testnetHashToken._utils.isValidValidatorAddress(''), false);
        });
    });
    describe('Recover transaction: success path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = hash_1.wrwUser.destinationAddress;
        const coin = statics_1.coins.get('thash:ylds');
        const testBalance = [
            {
                denom: 'nhash',
                amount: '150000000000',
            },
            {
                denom: 'uylds.fcc',
                amount: '10000',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.HashToken.prototype, 'getAccountBalance');
            accountBalance.withArgs(hash_1.wrwUser.senderAddress).resolves(testBalance);
            const accountDetails = sandBox.stub(src_1.HashToken.prototype, 'getAccountDetails');
            accountDetails.withArgs(hash_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.HashToken.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should recover funds for non-bitgo recoveries', async function () {
            const res = await testnetHashToken.recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(testnetHashToken.getAccountBalance);
            sandBox.assert.calledOnce(testnetHashToken.getAccountDetails);
            sandBox.assert.calledOnce(testnetHashToken.getChainId);
            const txn = new abstract_cosmos_1.CosmosTransaction(coin, testnetUtils);
            txn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const txnJson = txn.toJson();
            // The last sendMessage is of the native coin type
            const len = txnJson.sendMessages.length;
            const sendMessage = txnJson.sendMessages[len - 1].value;
            const tokenSendMessage = txnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const gasAmount = new bignumber_js_1.default(5000000000);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.toAddress, destinationAddress);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
            should.equal(sendMessage.amount[0].denom, testBalance[0].denom);
            should.equal(tokenSendMessage.toAddress, destinationAddress);
            should.equal(tokenSendMessage.amount[0].amount, testBalance[1].amount);
            should.equal(tokenSendMessage.amount[0].denom, testBalance[1].denom);
        });
        it('should recover funds for Unsigned Sweep Transaction', async function () {
            const res = await testnetHashToken.recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(testnetHashToken.getAccountBalance);
            sandBox.assert.calledOnce(testnetHashToken.getAccountDetails);
            sandBox.assert.calledOnce(testnetHashToken.getChainId);
            const unsignedSweepTxnDeserialize = new abstract_cosmos_1.CosmosTransaction(coin, testnetUtils);
            unsignedSweepTxnDeserialize.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const unsignedSweepTxnJson = unsignedSweepTxnDeserialize.toJson();
            // The last sendMessage is of the native coin type
            const len = unsignedSweepTxnJson.sendMessages.length;
            const sendMessage = unsignedSweepTxnJson.sendMessages[len - 1].value;
            const tokenSendMessage = unsignedSweepTxnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const gasAmount = new bignumber_js_1.default(5000000000);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.toAddress, destinationAddress);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
            should.equal(sendMessage.amount[0].denom, testBalance[0].denom);
            should.equal(tokenSendMessage.toAddress, destinationAddress);
            should.equal(tokenSendMessage.amount[0].amount, testBalance[1].amount);
            should.equal(tokenSendMessage.amount[0].denom, testBalance[1].denom);
        });
    });
    describe('Recover transaction: failure path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = hash_1.wrwUser.destinationAddress;
        const testZeroBalance = [
            {
                denom: 'nhash',
                amount: '0',
            },
            {
                denom: 'uylds.fcc',
                amount: '1000',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.HashToken.prototype, 'getAccountBalance');
            accountBalance.withArgs(hash_1.wrwUser.senderAddress).resolves(testZeroBalance);
            const accountDetails = sandBox.stub(src_1.HashToken.prototype, 'getAccountDetails');
            accountDetails.withArgs(hash_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.HashToken.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should throw error if backupkey is not present', async function () {
            await testnetHashToken
                .recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing backupKey');
        });
        it('should throw error if userkey is not present', async function () {
            await testnetHashToken
                .recover({
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing userKey');
        });
        it('should throw error if wallet passphrase is not present', async function () {
            await testnetHashToken
                .recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing wallet passphrase');
        });
        it('should throw error if there is no balance', async function () {
            await testnetHashToken
                .recover({
                userKey: hash_1.wrwUser.userPrivateKey,
                backupKey: hash_1.wrwUser.backupPrivateKey,
                bitgoKey: hash_1.wrwUser.bitgoPublicKey,
                walletPassphrase: hash_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('Did not have enough funds to recover');
        });
    });
});
