// @bitgo/sdk-lib-mpc/dist/src/curves/secp256k1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Secp256k1Curve = void 0;
const util_1 = require("../util");
const secp256k1_1 = require("@noble/curves/secp256k1");
const modular_1 = require("@noble/curves/abstract/modular");
const order = secp256k1_1.secp256k1.CURVE.n;
const privateKeySize = 32;
const publicKeySize = 33;
class Secp256k1Curve {
    constructor() {
        this.scalarBytes = privateKeySize;
        // Always use compressed points.
        this.pointBytes = publicKeySize;
    }
    scalarRandom() {
        return (0, util_1.bigIntFromU8ABE)(secp256k1_1.secp256k1.utils.randomPrivateKey());
    }
    scalarAdd(x, y) {
        return (0, modular_1.mod)(x + y, order);
    }
    scalarSub(x, y) {
        const negatedY = order - y;
        return (0, modular_1.mod)(x + negatedY, order);
    }
    scalarMult(x, y) {
        return (0, modular_1.mod)(x * y, order);
    }
    scalarReduce(s) {
        return (0, modular_1.mod)(s, order);
    }
    scalarNegate(s) {
        return order - s;
    }
    scalarInvert(s) {
        return (0, modular_1.invert)(s, order);
    }
    pointAdd(a, b) {
        const pointA = secp256k1_1.secp256k1.ProjectivePoint.fromHex((0, util_1.bigIntToBufferBE)(a, privateKeySize));
        const pointB = secp256k1_1.secp256k1.ProjectivePoint.fromHex((0, util_1.bigIntToBufferBE)(b, privateKeySize));
        return (0, util_1.bigIntFromU8ABE)(pointA.add(pointB).toRawBytes(true));
    }
    pointMultiply(p, s) {
        const pointA = secp256k1_1.secp256k1.ProjectivePoint.fromHex((0, util_1.bigIntToBufferBE)(p, privateKeySize));
        return (0, util_1.bigIntFromU8ABE)(pointA.multiply(s).toRawBytes(true));
    }
    basePointMult(n) {
        const point = (0, util_1.bigIntToBufferBE)(n, privateKeySize);
        return (0, util_1.bigIntFromU8ABE)(secp256k1_1.secp256k1.getPublicKey(point, true));
    }
    verify(message, signature, publicKey) {
        const sig = secp256k1_1.secp256k1.Signature.fromCompact(Buffer.from(signature.subarray(1))).addRecoveryBit(signature[0]);
        const pubFromSig = sig.recoverPublicKey(message).toRawBytes(true);
        return Buffer.from(pubFromSig).equals((0, util_1.bigIntToBufferBE)(publicKey, publicKeySize));
    }
    order() {
        return order;
    }
}
exports.Secp256k1Curve = Secp256k1Curve;
