// @bitgo/sdk-coin-sol/dist/cjs/test/unit/transactionBuilder/customInstructionBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const web3_js_1 = require("@solana/web3.js");
const getBuilderFactory_1 = require("../getBuilderFactory");
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/sol"));
const bs58_1 = __importDefault(require("bs58"));
describe('Sol Custom Instruction Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsol');
    const customInstructionBuilder = () => {
        const txBuilder = factory.getCustomInstructionBuilder();
        txBuilder.nonce(recentBlockHash);
        txBuilder.sender(authAccount.pub);
        return txBuilder;
    };
    const authAccount = new src_1.KeyPair(testData.authAccount).getKeys();
    const otherAccount = new src_1.KeyPair({ prv: testData.prvKeys.prvKey1.base58 }).getKeys();
    const recentBlockHash = 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi';
    const memo = 'test memo';
    // Helper function to convert TransactionInstruction to the expected format
    const convertInstructionToParams = (instruction) => ({
        programId: instruction.programId.toString(),
        keys: instruction.keys.map((key) => ({
            pubkey: key.pubkey.toString(),
            isSigner: key.isSigner,
            isWritable: key.isWritable,
        })),
        data: instruction.data.toString('hex'),
    });
    describe('Succeed', () => {
        it('build a transaction with a single custom instruction', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const txBuilder = customInstructionBuilder();
            txBuilder.addCustomInstruction(convertInstructionToParams(transferInstruction));
            const tx = await txBuilder.build();
            tx.inputs.length.should.equal(0);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
        });
        it('build a transaction with multiple custom instructions', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const priorityFeeInstruction = web3_js_1.ComputeBudgetProgram.setComputeUnitPrice({
                microLamports: 1000,
            });
            const txBuilder = customInstructionBuilder();
            txBuilder.addCustomInstructions([
                convertInstructionToParams(transferInstruction),
                convertInstructionToParams(priorityFeeInstruction),
            ]);
            const tx = await txBuilder.build();
            tx.inputs.length.should.equal(0);
            tx.outputs.length.should.equal(0);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            // Should have 2 instructions
            tx.solTransaction.instructions.should.have.length(2);
        });
        it('build a transaction with custom instruction and memo', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const txBuilder = customInstructionBuilder();
            txBuilder.addCustomInstruction(convertInstructionToParams(transferInstruction));
            txBuilder.memo(memo);
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            // Should have instruction + memo
            tx.solTransaction.instructions.should.have.length(2);
        });
        it('build a signed transaction with custom instruction', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const txBuilder = customInstructionBuilder();
            txBuilder.addCustomInstruction(convertInstructionToParams(transferInstruction));
            txBuilder.sign({ key: authAccount.prv });
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            // Should be signed
            tx.solTransaction.signatures.should.not.be.empty();
        });
        it('clear instructions from builder', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const txBuilder = customInstructionBuilder();
            txBuilder.addCustomInstruction(convertInstructionToParams(transferInstruction));
            txBuilder.getInstructions().should.have.length(1);
            txBuilder.clearInstructions();
            txBuilder.getInstructions().should.have.length(0);
        });
    });
    describe('Fail', () => {
        it('for null instruction', () => {
            const txBuilder = customInstructionBuilder();
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (0, should_1.default)(() => txBuilder.addCustomInstruction(null)).throwError('Instruction cannot be null or undefined');
        });
        it('for undefined instruction', () => {
            const txBuilder = customInstructionBuilder();
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (0, should_1.default)(() => txBuilder.addCustomInstruction(undefined)).throwError('Instruction cannot be null or undefined');
        });
        it('for instruction without programId', () => {
            const txBuilder = customInstructionBuilder();
            const invalidInstruction = {
                keys: [],
                data: '',
            };
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (0, should_1.default)(() => txBuilder.addCustomInstruction(invalidInstruction)).throwError('Versioned instruction must have a valid programIdIndex number');
        });
        it('for instruction without keys', () => {
            const txBuilder = customInstructionBuilder();
            const invalidInstruction = {
                programId: '11111111111111111111111111111112',
                data: '',
            };
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (0, should_1.default)(() => txBuilder.addCustomInstruction(invalidInstruction)).throwError('Instruction must have valid keys array');
        });
        it('for instruction without data', () => {
            const txBuilder = customInstructionBuilder();
            const invalidInstruction = {
                programId: '11111111111111111111111111111112',
                keys: [],
            };
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (0, should_1.default)(() => txBuilder.addCustomInstruction(invalidInstruction)).throwError('Instruction must have valid data string');
        });
        it('for non-array in addCustomInstructions', () => {
            const txBuilder = customInstructionBuilder();
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (0, should_1.default)(() => txBuilder.addCustomInstructions('invalid')).throwError('Instructions must be an array');
        });
        it('when building without instructions', async () => {
            const txBuilder = customInstructionBuilder();
            await txBuilder.build().should.be.rejectedWith('At least one custom instruction must be specified');
        });
        it('when building without sender', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const txBuilder = factory.getCustomInstructionBuilder();
            txBuilder.nonce(recentBlockHash);
            txBuilder.addCustomInstruction(convertInstructionToParams(transferInstruction));
            await txBuilder.build().should.be.rejectedWith('Invalid transaction: missing sender');
        });
        it('when building without nonce', async () => {
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            });
            const txBuilder = factory.getCustomInstructionBuilder();
            txBuilder.sender(authAccount.pub);
            txBuilder.addCustomInstruction(convertInstructionToParams(transferInstruction));
            await txBuilder.build().should.be.rejectedWith('Invalid transaction: missing nonce blockhash');
        });
    });
    describe('SolVersionedInstruction Support', () => {
        it('should accept and validate SolVersionedInstruction format', () => {
            const compiledInstruction = {
                programIdIndex: 1,
                accountKeyIndexes: [0, 2, 3],
                data: '3Bxs43ZMjSRQLs6o', // base58 encoded instruction data
            };
            const txBuilder = customInstructionBuilder();
            (0, should_1.default)(() => txBuilder.addCustomInstruction(compiledInstruction)).not.throwError();
            txBuilder.getInstructions().should.have.length(1);
            const addedInstruction = txBuilder.getInstructions()[0];
            addedInstruction.params.should.deepEqual(compiledInstruction);
        });
        it('should validate versioned instruction format', () => {
            const txBuilder = customInstructionBuilder();
            // Invalid programIdIndex
            (0, should_1.default)(() => txBuilder.addCustomInstruction({
                programIdIndex: -1,
                accountKeyIndexes: [0],
                data: '1', // base58 for 0x00
            })).throwError('Versioned instruction must have a valid programIdIndex number');
            // Invalid accountKeyIndexes
            (0, should_1.default)(() => txBuilder.addCustomInstruction({
                programIdIndex: 0,
                accountKeyIndexes: [-1],
                data: '1', // base58 for 0x00
            })).throwError('Each accountKeyIndex must be a non-negative number');
            // Missing data
            (0, should_1.default)(() => txBuilder.addCustomInstruction({
                programIdIndex: 0,
                accountKeyIndexes: [0],
            })).throwError('Versioned instruction must have valid data string');
        });
        it('should handle mixed SolInstruction and SolVersionedInstruction', () => {
            const traditionalInstruction = convertInstructionToParams(web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(authAccount.pub),
                toPubkey: new web3_js_1.PublicKey(otherAccount.pub),
                lamports: 1000000,
            }));
            const compiledInstruction = {
                programIdIndex: 1,
                accountKeyIndexes: [0, 2],
                data: '4HSo5YVBrgChTZX5',
            };
            const txBuilder = customInstructionBuilder();
            txBuilder.addCustomInstruction(traditionalInstruction);
            txBuilder.addCustomInstruction(compiledInstruction);
            txBuilder.getInstructions().should.have.length(2);
        });
    });
    describe('fromVersionedTransactionData', () => {
        function extractVersionedTransactionData(base64Bytes) {
            const buffer = Buffer.from(base64Bytes, 'base64');
            const versionedTx = web3_js_1.VersionedTransaction.deserialize(buffer);
            return {
                versionedInstructions: versionedTx.message.compiledInstructions.map((ci) => ({
                    programIdIndex: ci.programIdIndex,
                    accountKeyIndexes: ci.accountKeyIndexes,
                    data: bs58_1.default.encode(ci.data),
                })),
                addressLookupTables: versionedTx.message.addressTableLookups.map((alt) => ({
                    accountKey: alt.accountKey.toString(),
                    writableIndexes: alt.writableIndexes,
                    readonlyIndexes: alt.readonlyIndexes,
                })),
                staticAccountKeys: versionedTx.message.staticAccountKeys.map((key) => key.toString()),
                messageHeader: {
                    numRequiredSignatures: versionedTx.message.header.numRequiredSignatures,
                    numReadonlySignedAccounts: versionedTx.message.header.numReadonlySignedAccounts,
                    numReadonlyUnsignedAccounts: versionedTx.message.header.numReadonlyUnsignedAccounts,
                },
            };
        }
        it('should process VersionedTransactionData and extract instructions', () => {
            const txBuilder = customInstructionBuilder();
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            // Should parse without throwing errors
            (0, should_1.default)(() => txBuilder.fromVersionedTransactionData(versionedTxData)).not.throwError();
            // Should have extracted instructions
            const instructions = txBuilder.getInstructions();
            instructions.length.should.be.greaterThan(0);
            for (const instruction of instructions) {
                instruction.type.should.equal('VersionedCustomInstruction');
                instruction.params.should.have.property('programIdIndex');
                instruction.params.should.have.property('accountKeyIndexes');
                instruction.params.should.have.property('data');
            }
        });
        it('should store VersionedTransactionData in underlying transaction', async () => {
            const txBuilder = customInstructionBuilder();
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            txBuilder.fromVersionedTransactionData(versionedTxData);
            const tx = txBuilder['_transaction'];
            const storedData = tx.getVersionedTransactionData();
            should_1.default.exist(storedData);
            storedData.versionedInstructions.length.should.equal(versionedTxData.versionedInstructions.length);
            storedData.addressLookupTables.length.should.equal(versionedTxData.addressLookupTables.length);
            storedData.staticAccountKeys.length.should.equal(versionedTxData.staticAccountKeys.length);
        });
        it('should validate input data', () => {
            const txBuilder = customInstructionBuilder();
            // Invalid: null/undefined
            (0, should_1.default)(() => txBuilder.fromVersionedTransactionData(null)).throwError(/must be a valid object/);
            (0, should_1.default)(() => txBuilder.fromVersionedTransactionData(undefined)).throwError(/must be a valid object/);
            // Invalid: empty instructions
            (0, should_1.default)(() => txBuilder.fromVersionedTransactionData({
                versionedInstructions: [],
                addressLookupTables: [],
                staticAccountKeys: ['test'],
                messageHeader: { numRequiredSignatures: 1, numReadonlySignedAccounts: 0, numReadonlyUnsignedAccounts: 0 },
            })).throwError(/non-empty array/);
            // Invalid: missing addressLookupTables
            (0, should_1.default)(() => txBuilder.fromVersionedTransactionData({
                versionedInstructions: [{ programIdIndex: 0, accountKeyIndexes: [0], data: '1' }], // base58 for 0x00
                staticAccountKeys: ['test'],
            })).throwError(/must be an array/);
            // Invalid: empty staticAccountKeys
            (0, should_1.default)(() => txBuilder.fromVersionedTransactionData({
                versionedInstructions: [{ programIdIndex: 0, accountKeyIndexes: [0], data: '1' }], // base58 for 0x00
                addressLookupTables: [],
                staticAccountKeys: [],
                messageHeader: { numRequiredSignatures: 1, numReadonlySignedAccounts: 0, numReadonlyUnsignedAccounts: 0 },
            })).throwError(/non-empty array/);
        });
        it('should work with the complete transaction building flow', async () => {
            const txBuilder = customInstructionBuilder();
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            // Process the VersionedTransactionData
            txBuilder.fromVersionedTransactionData(versionedTxData);
            txBuilder.sender(versionedTxData.staticAccountKeys[0]); // Fee payer
            txBuilder.nonce(testData.blockHashes.validBlockHashes[0]);
            const tx = await txBuilder.build();
            // Verify transaction properties
            tx.should.be.ok();
            tx.type.should.equal(31); // TransactionType.CustomTx enum value
            // Verify signable payload
            const payload = tx.signablePayload;
            payload.should.be.instanceOf(Buffer);
            payload.length.should.be.greaterThan(0);
            // Verify payload is deterministic - rebuilding with same params produces same payload
            const txBuilder2 = customInstructionBuilder();
            txBuilder2.fromVersionedTransactionData(versionedTxData);
            txBuilder2.sender(versionedTxData.staticAccountKeys[0]);
            txBuilder2.nonce(testData.blockHashes.validBlockHashes[0]);
            const tx2 = await txBuilder2.build();
            should_1.default.equal(tx2.signablePayload.toString('hex'), payload.toString('hex'));
        });
        it('should extract the correct number of instructions from Jupiter transaction', () => {
            const txBuilder = customInstructionBuilder();
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            txBuilder.fromVersionedTransactionData(versionedTxData);
            const instructions = txBuilder.getInstructions();
            // Verify we extracted instructions
            instructions.length.should.be.greaterThan(0);
            instructions.length.should.equal(versionedTxData.versionedInstructions.length);
            // Verify each instruction has valid compiled format
            for (const instruction of instructions) {
                const params = instruction.params;
                (typeof params.programIdIndex).should.equal('number');
                params.programIdIndex.should.be.greaterThanOrEqual(0);
                Array.isArray(params.accountKeyIndexes).should.equal(true);
                (typeof params.data).should.equal('string');
                // Data should be valid base58 string (non-empty)
                params.data.length.should.be.greaterThan(0);
                // Verify it can be decoded
                should_1.default.doesNotThrow(() => bs58_1.default.decode(params.data));
            }
        });
        it('should clear instructions properly after using fromVersionedTransactionData', () => {
            const txBuilder = customInstructionBuilder();
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            txBuilder.fromVersionedTransactionData(versionedTxData);
            txBuilder.getInstructions().length.should.be.greaterThan(0);
            // Clear should work
            txBuilder.clearInstructions();
            txBuilder.getInstructions().should.have.length(0);
        });
        it('should extract fee payer from staticAccountKeys', () => {
            const txBuilder = factory.getCustomInstructionBuilder();
            txBuilder.nonce(recentBlockHash);
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            txBuilder.fromVersionedTransactionData(versionedTxData);
            const sender = txBuilder['_sender'];
            should_1.default.exist(sender);
            sender.should.equal(versionedTxData.staticAccountKeys[0]);
        });
        it('should not override existing sender', () => {
            const txBuilder = customInstructionBuilder();
            const versionedTxData = extractVersionedTransactionData(testData.JUPITER_VERSIONED_TX_BYTES);
            const originalSender = txBuilder['_sender'];
            txBuilder.fromVersionedTransactionData(versionedTxData);
            const sender = txBuilder['_sender'];
            sender.should.equal(originalSender);
            sender.should.equal(authAccount.pub);
        });
    });
});
