// @bitgo/sdk-coin-sol/dist/cjs/test/unit/transactionBuilder/stakingActivateBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const testData = __importStar(require("../../resources/sol"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const src_1 = require("../../../src");
const constants_1 = require("../../../src/lib/constants");
const public_types_1 = require("@bitgo/public-types");
describe('Sol Staking Activate Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsol');
    const stakingBuilder = () => {
        const txBuilder = factory.getStakingActivateBuilder();
        txBuilder.nonce(recentBlockHash);
        txBuilder.sender(wallet.pub);
        return txBuilder;
    };
    // not valid data
    const invalidPubKey = testData.pubKeys.invalidPubKeys[0];
    const wrongAccount = new src_1.KeyPair({ prv: testData.prvKeys.prvKey1.base58 }).getKeys();
    // valid data
    const wallet = new src_1.KeyPair(testData.authAccount).getKeys();
    const stakeAccount = new src_1.KeyPair(testData.stakeAccount).getKeys();
    const validator = testData.validator;
    const recentBlockHash = 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi';
    const amount = '300000';
    const performTest = async ({ signBuilder, knownRawTx, makeUnsignedBuilder, verifyBuiltTransaction, }) => {
        // Build transaction
        const txBuilder = makeUnsignedBuilder();
        const unsignedTx = await txBuilder.build();
        const tx = signBuilder ? await signBuilder(txBuilder).build() : unsignedTx;
        // Verify built transaction
        verifyBuiltTransaction(tx);
        // Verify raw transaction
        const rawTx = tx.toBroadcastFormat();
        should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
        should_1.default.equal(rawTx, knownRawTx);
        // Rebuild transaction and verify
        const builderFromRawTx = factory.from(rawTx);
        const rebuiltTx = await builderFromRawTx.build();
        should_1.default.equal(rebuiltTx.toBroadcastFormat(), unsignedTx.toBroadcastFormat());
        should_1.default.equal(rebuiltTx.signablePayload.toString('hex'), unsignedTx.signablePayload.toString('hex'));
        should_1.default.deepEqual(rebuiltTx.toJson().instructionsData, tx.toJson().instructionsData);
    };
    const makeUnsignedBuilderNative = (doMemo) => {
        const txBuilder = factory.getStakingActivateBuilder();
        txBuilder
            .amount(amount)
            .sender(wallet.pub)
            .stakingAddress(stakeAccount.pub)
            .validator(validator.pub)
            .nonce(recentBlockHash);
        if (doMemo) {
            txBuilder.memo('test memo');
        }
        return txBuilder;
    };
    const verifyBuiltTransactionNativeOrMarinade = (tx, doMemo, stakingType) => {
        const txJson = tx.toJson();
        txJson.instructionsData.should.deepEqual([
            ...(doMemo
                ? [
                    {
                        type: 'Memo',
                        params: {
                            memo: 'test memo',
                        },
                    },
                ]
                : []),
            {
                type: 'Activate',
                params: {
                    fromAddress: wallet.pub,
                    stakingAddress: stakeAccount.pub,
                    amount: amount,
                    validator: validator.pub,
                    stakingType,
                },
            },
        ]);
        tx.inputs.should.deepEqual([
            {
                address: wallet.pub,
                value: amount,
                coin: 'tsol',
            },
        ]);
        tx.outputs.should.deepEqual([
            {
                address: stakeAccount.pub,
                value: amount,
                coin: 'tsol',
            },
        ]);
    };
    const signBuilderNativeOrMarinade = (txBuilder) => {
        txBuilder.sign({ key: wallet.prv });
        txBuilder.sign({ key: stakeAccount.prv });
        return txBuilder;
    };
    const makeUnsignedBuilderMarinade = (doMemo) => {
        return makeUnsignedBuilderNative(doMemo).stakingType(public_types_1.SolStakingTypeEnum.MARINADE);
    };
    const makeUnsignedBuilderJito = (doMemo, doCreateATA) => {
        const txBuilder = factory.getStakingActivateBuilder();
        txBuilder
            .amount(amount)
            .sender(wallet.pub)
            .stakingAddress(constants_1.JITO_STAKE_POOL_ADDRESS)
            .validator(constants_1.JITO_STAKE_POOL_ADDRESS)
            .stakingType(public_types_1.SolStakingTypeEnum.JITO)
            .extraParams({
            stakePoolData: {
                managerFeeAccount: testData.JITO_STAKE_POOL_DATA_PARSED.managerFeeAccount.toString(),
                poolMint: testData.JITO_STAKE_POOL_DATA_PARSED.poolMint.toString(),
                reserveStake: testData.JITO_STAKE_POOL_DATA_PARSED.reserveStake.toString(),
            },
            createAssociatedTokenAccount: doCreateATA,
        })
            .nonce(recentBlockHash);
        if (doMemo) {
            txBuilder.memo('test memo');
        }
        return txBuilder;
    };
    const signBuilderJito = (txBuilder) => {
        txBuilder.sign({ key: wallet.prv });
        return txBuilder;
    };
    const verifyBuiltTransactionJito = (tx, doMemo, doCreateATA) => {
        const txJson = tx.toJson();
        const expectedInstructions = [];
        if (doCreateATA) {
            expectedInstructions.push({
                type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                params: {
                    ataAddress: '2vJrx2Bn7PifLZDRaSCpphE9WtZsx1k43SRyiQDhE1As',
                    mintAddress: constants_1.JITOSOL_MINT_ADDRESS,
                    ownerAddress: wallet.pub,
                    payerAddress: wallet.pub,
                    tokenName: 'sol:jitosol',
                },
            });
        }
        if (doMemo) {
            expectedInstructions.push({
                type: constants_1.InstructionBuilderTypes.Memo,
                params: {
                    memo: 'test memo',
                },
            });
        }
        expectedInstructions.push({
            type: constants_1.InstructionBuilderTypes.StakingActivate,
            params: {
                fromAddress: wallet.pub,
                stakingAddress: constants_1.JITO_STAKE_POOL_ADDRESS,
                amount: amount,
                validator: constants_1.JITO_STAKE_POOL_ADDRESS,
                stakingType: public_types_1.SolStakingTypeEnum.JITO,
                extraParams: {
                    stakePoolData: {
                        managerFeeAccount: testData.JITO_STAKE_POOL_DATA_PARSED.managerFeeAccount.toString(),
                        poolMint: testData.JITO_STAKE_POOL_DATA_PARSED.poolMint.toString(),
                        reserveStake: testData.JITO_STAKE_POOL_DATA_PARSED.reserveStake.toString(),
                    },
                    createAssociatedTokenAccount: doCreateATA,
                },
            },
        });
        txJson.instructionsData.should.deepEqual(expectedInstructions);
        tx.inputs.should.deepEqual([
            {
                address: wallet.pub,
                value: amount,
                coin: 'tsol',
            },
        ]);
        tx.outputs.should.deepEqual([]);
    };
    describe('Succeed', () => {
        describe('Native staking tests', () => {
            const performTestNative = async (doMemo, doSign, knownRawTx) => {
                await performTest({
                    signBuilder: doSign ? signBuilderNativeOrMarinade : undefined,
                    knownRawTx,
                    makeUnsignedBuilder: () => makeUnsignedBuilderNative(doMemo),
                    verifyBuiltTransaction: (x) => verifyBuiltTransactionNativeOrMarinade(x, doMemo, public_types_1.SolStakingTypeEnum.NATIVE),
                });
            };
            it('build a create and delegate staking signed tx', async () => performTestNative(false, true, testData.STAKING_ACTIVATE_SIGNED_TX));
            it('build a create and delegate staking signed tx with memo', async () => performTestNative(true, true, testData.STAKING_ACTIVATE_SIGNED_TX_WITH_MEMO));
            it('build a create and delegate staking unsigned tx', async () => performTestNative(false, false, testData.STAKING_ACTIVATE_UNSIGNED_TX));
            it('build a create and delegate staking unsigned tx with memo', async () => performTestNative(true, false, testData.STAKING_ACTIVATE_UNSIGNED_TX_WITH_MEMO));
        });
        describe('Marinade staking tests', () => {
            const performTestMarinade = async (doMemo, doSign, knownRawTx) => {
                await performTest({
                    signBuilder: doSign ? signBuilderNativeOrMarinade : undefined,
                    knownRawTx,
                    makeUnsignedBuilder: () => makeUnsignedBuilderMarinade(doMemo),
                    verifyBuiltTransaction: (x) => verifyBuiltTransactionNativeOrMarinade(x, doMemo, public_types_1.SolStakingTypeEnum.MARINADE),
                });
            };
            it('build a create staking signed tx', async () => performTestMarinade(false, true, testData.MARINADE_STAKING_ACTIVATE_SIGNED_TX));
            it('build a create signed tx with memo', async () => performTestMarinade(true, true, testData.MARINADE_STAKING_ACTIVATE_SIGNED_TX_WITH_MEMO));
            it('build a create unsigned tx', async () => performTestMarinade(false, false, testData.MARINADE_STAKING_ACTIVATE_UNSIGNED_TX));
            it('build a create unsigned tx with memo', async () => performTestMarinade(true, false, testData.MARINADE_STAKING_ACTIVATE_UNSIGNED_TX_WITH_MEMO));
        });
        describe('Jito staking tests', () => {
            const performTestJito = async (doMemo, doSign, doCreateATA, knownRawTx) => {
                await performTest({
                    signBuilder: doSign ? signBuilderJito : undefined,
                    knownRawTx,
                    makeUnsignedBuilder: () => makeUnsignedBuilderJito(doMemo, doCreateATA),
                    verifyBuiltTransaction: (x) => verifyBuiltTransactionJito(x, doMemo, doCreateATA),
                });
            };
            const MEMO = 1;
            const SIGN = 2;
            const ATA = 4;
            const rawTxData = Array.from(new Array(8)).map(() => undefined);
            rawTxData[0] = testData.JITO_STAKING_ACTIVATE_UNSIGNED_TX;
            rawTxData[MEMO] = testData.JITO_STAKING_ACTIVATE_UNSIGNED_TX_WITH_MEMO;
            rawTxData[SIGN] = testData.JITO_STAKING_ACTIVATE_SIGNED_TX;
            rawTxData[MEMO + SIGN] = testData.JITO_STAKING_ACTIVATE_SIGNED_TX_WITH_MEMO;
            rawTxData[ATA] = testData.JITO_STAKING_ACTIVATE_UNSIGNED_TX_WITH_ATA;
            rawTxData[ATA + MEMO] = testData.JITO_STAKING_ACTIVATE_UNSIGNED_TX_WITH_MEMO_WITH_ATA;
            rawTxData[ATA + SIGN] = testData.JITO_STAKING_ACTIVATE_SIGNED_TX_WITH_ATA;
            rawTxData[ATA + MEMO + SIGN] = testData.JITO_STAKING_ACTIVATE_SIGNED_TX_WITH_MEMO_WITH_ATA;
            for (const doMemo of [false, true]) {
                for (const doSign of [false, true]) {
                    for (const doCreateATA of [false, true]) {
                        it(`build a tx with doMemo=${doMemo}, doSign=${doSign}, doCreateATA=${doCreateATA}`, async () => {
                            const data = rawTxData[(doMemo ? MEMO : 0) + (doSign ? SIGN : 0) + (doCreateATA ? ATA : 0)];
                            (0, should_1.default)(data).not.be.undefined();
                            await performTestJito(doMemo, doSign, doCreateATA, data ?? '');
                        });
                    }
                }
            }
        });
    });
    describe('Fail', () => {
        it('for invalid sender address', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.sender(invalidPubKey)).throwError('Invalid or missing sender, got: ' + invalidPubKey);
        });
        it('for invalid staking address', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.stakingAddress(invalidPubKey)).throwError('Invalid or missing stakingAddress, got: ' + invalidPubKey);
        });
        it('for invalid validator address', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.validator(invalidPubKey)).throwError('Invalid or missing validator, got: ' + invalidPubKey);
        });
        it('build a staking activate tx when amount is invalid', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.amount('randomstring')).throwError('Value cannot be zero or less');
        });
        it('build a staking activate tx when amount is less than zero', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.amount('-1')).throwError('Value cannot be zero or less');
        });
        it('build a staking activate tx when amount is equal to zero', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.amount('0')).throwError('Value cannot be zero or less');
        });
        it('build a staking activate tx and sign with an incorrect account', async () => {
            const txBuilder = stakingBuilder();
            txBuilder.sender(wallet.pub);
            txBuilder.stakingAddress(stakeAccount.pub);
            txBuilder.validator(validator.pub);
            txBuilder.amount(amount);
            txBuilder.sign({ key: wrongAccount.prv });
            await txBuilder.build().should.rejectedWith('unknown signer: CP5Dpaa42RtJmMuKqCQsLwma5Yh3knuvKsYDFX85F41S');
        });
        it('build a staking activate tx with the same sender and staking address', async () => {
            const txBuilder = stakingBuilder();
            txBuilder.sender(wallet.pub);
            txBuilder.stakingAddress(wallet.pub);
            txBuilder.validator(validator.pub);
            txBuilder.amount(amount);
            txBuilder.sign({ key: wrongAccount.prv });
            await txBuilder.build().should.rejectedWith('Sender address cannot be the same as the Staking address');
        });
        it('build when nonce is not provided', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder.sender(wallet.pub);
            txBuilder.stakingAddress(stakeAccount.pub);
            txBuilder.amount(amount);
            txBuilder.validator(validator.pub);
            txBuilder.sign({ key: wallet.prv });
            await txBuilder.build().should.rejectedWith('Invalid transaction: missing nonce blockhash');
        });
        it('build when sender is not provided', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder.stakingAddress(stakeAccount.pub);
            txBuilder.amount(amount);
            txBuilder.validator(validator.pub);
            txBuilder.nonce(recentBlockHash);
            txBuilder.sign({ key: wallet.prv });
            await txBuilder.build().should.rejectedWith('Invalid transaction: missing sender');
        });
        it('build when stakingAddress is not provided', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder.sender(wallet.pub);
            txBuilder.amount(amount);
            txBuilder.validator(validator.pub);
            txBuilder.nonce(recentBlockHash);
            txBuilder.sign({ key: wallet.prv });
            await txBuilder.build().should.rejectedWith('Staking Address must be set before building the transaction');
        });
        it('build when validator is not provided', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder.stakingAddress(stakeAccount.pub);
            txBuilder.sender(wallet.pub);
            txBuilder.amount(amount);
            txBuilder.nonce(recentBlockHash);
            txBuilder.sign({ key: wallet.prv });
            await txBuilder.build().should.rejectedWith('Validator must be set before building the transaction');
        });
        it('to sign twice with the same key', () => {
            const txBuilder = factory.from(testData.STAKING_ACTIVATE_UNSIGNED_TX);
            txBuilder.sign({ key: wallet.prv });
            (0, should_1.default)(() => txBuilder.sign({ key: wallet.prv })).throwError('Duplicated signer: ' + wallet.prv?.toString());
        });
    });
    describe('From and sign', () => {
        describe('Succeed', () => {
            it('build from an unsigned staking activate and sign it', async () => {
                const txBuilder = factory.from(testData.STAKING_ACTIVATE_UNSIGNED_TX);
                txBuilder.sign({ key: wallet.prv });
                txBuilder.sign({ key: stakeAccount.prv });
                const tx = await txBuilder.build();
                tx.inputs.length.should.equal(1);
                tx.inputs[0].should.deepEqual({
                    address: wallet.pub,
                    value: amount,
                    coin: 'tsol',
                });
                tx.outputs.length.should.equal(1);
                const rawTx = tx.toBroadcastFormat();
                should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
                should_1.default.equal(rawTx, testData.STAKING_ACTIVATE_SIGNED_TX);
            });
            it('Marinade: build from an unsigned staking activate and sign it', async () => {
                const txBuilder = factory.from(testData.MARINADE_STAKING_ACTIVATE_UNSIGNED_TX);
                txBuilder.sign({ key: wallet.prv });
                txBuilder.sign({ key: stakeAccount.prv });
                const tx = await txBuilder.build();
                tx.inputs.length.should.equal(1);
                tx.inputs[0].should.deepEqual({
                    address: wallet.pub,
                    value: amount,
                    coin: 'tsol',
                });
                tx.outputs.length.should.equal(1);
                const rawTx = tx.toBroadcastFormat();
                should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
                should_1.default.equal(rawTx, testData.MARINADE_STAKING_ACTIVATE_SIGNED_TX);
            });
            it('build from an unsigned staking activate with memo and sign it', async () => {
                const txBuilder = factory.from(testData.STAKING_ACTIVATE_UNSIGNED_TX_WITH_MEMO);
                txBuilder.sign({ key: wallet.prv });
                txBuilder.sign({ key: stakeAccount.prv });
                const tx = await txBuilder.build();
                tx.inputs.length.should.equal(1);
                tx.inputs[0].should.deepEqual({
                    address: wallet.pub,
                    value: amount,
                    coin: 'tsol',
                });
                tx.outputs.length.should.equal(1);
                const rawTx = tx.toBroadcastFormat();
                should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
                should_1.default.equal(rawTx, testData.STAKING_ACTIVATE_SIGNED_TX_WITH_MEMO);
            });
        });
        describe('Fail', () => {
            it('build from an unsigned staking activate and fail to sign it', async () => {
                const txBuilder = factory.from(testData.STAKING_ACTIVATE_UNSIGNED_TX);
                txBuilder.sign({ key: wrongAccount.prv });
                await txBuilder.build().should.rejectedWith('unknown signer: CP5Dpaa42RtJmMuKqCQsLwma5Yh3knuvKsYDFX85F41S');
            });
            it('build from a signed staking activate and fail to sign it', async () => {
                const txBuilder = factory.from(testData.STAKING_ACTIVATE_SIGNED_TX);
                txBuilder.sign({ key: wrongAccount.prv });
                await txBuilder.build().should.rejectedWith('unknown signer: CP5Dpaa42RtJmMuKqCQsLwma5Yh3knuvKsYDFX85F41S');
            });
        });
    });
});
