// @bitgo/sdk-coin-sol/dist/cjs/test/unit/transactionBuilder/StakingWithdrawBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/sol"));
describe('Sol Staking Withdraw Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tsol');
    const stakingBuilder = () => {
        const txBuilder = factory.getStakingWithdrawBuilder();
        txBuilder.nonce(recentBlockHash);
        txBuilder.sender(wallet.pub);
        return txBuilder;
    };
    const wallet = new src_1.KeyPair(testData.authAccount).getKeys();
    const stakeAccount = new src_1.KeyPair(testData.stakeAccount).getKeys();
    const recentBlockHash = 'GHtXQBsoZHVnNFa9YevAzFr17DJjgHXk3ycTKD5xD3Zi';
    const amount = '300000';
    const invalidPubKey = testData.pubKeys.invalidPubKeys[0];
    describe('Should succeed', () => {
        it('building a staking withdraw tx', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder.sender(wallet.pub).stakingAddress(stakeAccount.pub).amount(amount).nonce(recentBlockHash);
            txBuilder.sign({ key: wallet.prv });
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            txJson.instructionsData.should.deepEqual([
                {
                    type: 'Withdraw',
                    params: {
                        fromAddress: wallet.pub,
                        stakingAddress: stakeAccount.pub,
                        amount: amount,
                    },
                },
            ]);
            should_1.default.equal(rawTx, testData.STAKING_WITHDRAW_SIGNED_TX);
        });
        it('building a staking withdraw signed tx with memo', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder
                .sender(wallet.pub)
                .stakingAddress(stakeAccount.pub)
                .amount(amount)
                .nonce(recentBlockHash)
                .memo('Test withdraw');
            txBuilder.sign({ key: wallet.prv });
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            const txJson = tx.toJson();
            txJson.instructionsData.should.deepEqual([
                {
                    type: 'Withdraw',
                    params: {
                        fromAddress: wallet.pub,
                        stakingAddress: stakeAccount.pub,
                        amount: amount,
                    },
                },
                {
                    type: 'Memo',
                    params: {
                        memo: 'Test withdraw',
                    },
                },
            ]);
            should_1.default.equal(rawTx, testData.STAKING_WITHDRAW_SIGNED_TX_WITH_MEMO);
        });
        it('building a staking withdraw unsigned tx', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder.sender(wallet.pub).stakingAddress(stakeAccount.pub).amount(amount).nonce(recentBlockHash);
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            const txJson = tx.toJson();
            txJson.instructionsData.should.deepEqual([
                {
                    type: 'Withdraw',
                    params: {
                        fromAddress: wallet.pub,
                        stakingAddress: stakeAccount.pub,
                        amount: amount,
                    },
                },
            ]);
            should_1.default.equal(rawTx, testData.STAKING_WITHDRAW_UNSIGNED_TX);
        });
        it('building a staking withdraw unsigned tx with memo', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder
                .sender(wallet.pub)
                .stakingAddress(stakeAccount.pub)
                .amount(amount)
                .nonce(recentBlockHash)
                .memo('Test withdraw');
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(src_1.Utils.isValidRawTransaction(rawTx), true);
            const txJson = tx.toJson();
            txJson.instructionsData.should.deepEqual([
                {
                    type: 'Withdraw',
                    params: {
                        fromAddress: wallet.pub,
                        stakingAddress: stakeAccount.pub,
                        amount: amount,
                    },
                },
                {
                    type: 'Memo',
                    params: {
                        memo: 'Test withdraw',
                    },
                },
            ]);
            should_1.default.equal(rawTx, testData.STAKING_WITHDRAW_UNSIGNED_TX_WITH_MEMO);
        });
        it('building an encoded unsigned transaction and signing it', async () => {
            const txBuilder = factory.from(testData.STAKING_WITHDRAW_UNSIGNED_TX_WITH_MEMO);
            txBuilder.sign({ key: wallet.prv });
            const tx = await txBuilder.build();
            should_1.default.equal(tx.toBroadcastFormat(), testData.STAKING_WITHDRAW_SIGNED_TX_WITH_MEMO);
        });
        it('building an encoded signed transaction', async () => {
            const txBuilder = factory.from(testData.STAKING_WITHDRAW_SIGNED_TX_WITH_MEMO);
            txBuilder.sign({ key: wallet.prv });
            const tx = await txBuilder.build();
            should_1.default.equal(tx.toBroadcastFormat(), testData.STAKING_WITHDRAW_SIGNED_TX_WITH_MEMO);
        });
    });
    describe('Should fail', () => {
        it('build a staking withdraw tx when amount is invalid', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.amount('randomstring')).throwError('Value cannot be zero or less');
        });
        it('build a staking withdraw tx when amount is less than zero', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.amount('-1')).throwError('Value cannot be zero or less');
        });
        it('build a staking withdraw tx when amount is equal to zero', () => {
            const txBuilder = stakingBuilder();
            (0, should_1.default)(() => txBuilder.amount('0')).throwError('Value cannot be zero or less');
        });
        it('building a staking withdraw tx without staking address', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder.sender(wallet.pub).nonce(recentBlockHash);
            txBuilder.sign({ key: wallet.prv });
            await txBuilder.build().should.be.rejectedWith('Staking address must be set before building the transaction');
        });
        it('building a staking withdraw tx with a wrong staking address', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder.sender(wallet.pub).nonce(recentBlockHash);
            (0, should_1.default)(() => txBuilder.stakingAddress(invalidPubKey)).throwError(`Invalid or missing stakingAddress, got: ${invalidPubKey}`);
        });
        it('building a staking withdraw tx with the same address as sender and staking', async () => {
            const txBuilder = factory.getStakingWithdrawBuilder();
            txBuilder.sender(wallet.pub).amount(amount).nonce(recentBlockHash).stakingAddress(wallet.pub);
            txBuilder.stakingAddress(wallet.pub);
            await txBuilder.build().should.rejectedWith('Sender address cannot be the same as the Staking address');
        });
    });
});
