// bitgo/dist/test/v2/unit/internal/tssUtils/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nockSendTxRequest = nockSendTxRequest;
exports.nockSendSignatureShare = nockSendSignatureShare;
exports.nockSendSignatureShareWithResponse = nockSendSignatureShareWithResponse;
exports.nockDeleteSignatureShare = nockDeleteSignatureShare;
exports.nockCreateTxRequest = nockCreateTxRequest;
exports.nockGetTxRequest = nockGetTxRequest;
exports.nockExchangeCommitments = nockExchangeCommitments;
exports.getRoute = getRoute;
const sdk_core_1 = require("@bitgo/sdk-core");
const nock = require("nock");
const _ = require("lodash");
async function nockSendTxRequest(params) {
    return nock('https://bitgo.fakeurl')
        .post(`/api/v2/${params.coin}/wallet/${params.walletId}/tx/send`, { txRequestId: params.txRequestId })
        .reply(200);
}
async function nockSendSignatureShare(params, status = 200) {
    params.requestType = params.requestType || sdk_core_1.RequestType.tx;
    params.apiMode = params.apiMode || 'lite';
    const transactions = getRoute(params.tssType, params.requestType, params.apiMode);
    return nock('https://bitgo.fakeurl')
        .persist(true)
        .post(`/api/v2/wallet/${params.walletId}/txrequests/${params.txRequestId + transactions}/signatureshares`)
        .reply(status, status === 200 ? params.signatureShare : { error: 'some error' });
}
async function nockSendSignatureShareWithResponse(params, status = 200) {
    const transactions = getRoute(params.tssType);
    return nock('https://bitgo.fakeurl')
        .post(`/api/v2/wallet/${params.walletId}/txrequests/${params.txRequestId + transactions}/signatureshares`)
        .reply(status, status === 200 ? params.response : { error: 'some error' });
}
async function nockDeleteSignatureShare(params, status = 200) {
    const transactions = getRoute(params.tssType);
    return nock('https://bitgo.fakeurl')
        .delete(`/api/v2/wallet/${params.walletId}/txrequests/${params.txRequestId + transactions}/signatureshares`)
        .reply(status, status === 200 ? [params.signatureShare] : { error: 'some error' });
}
async function nockCreateTxRequest(params) {
    return nock('https://bitgo.fakeurl')
        .post(`/api/v2/wallet/${params.walletId}/txrequests`, (body) => {
        return _.isEqual(body, params.requestBody);
    })
        .reply(200, params.response);
}
async function nockGetTxRequest(params) {
    return nock('https://bitgo.fakeurl')
        .persist(true)
        .get(`/api/v2/wallet/${params.walletId}/txrequests?txRequestIds=${params.txRequestId}&latest=true`)
        .reply(200, params.response);
}
async function nockExchangeCommitments(params) {
    const { apiMode = 'lite' } = params;
    let addendum = '';
    if (apiMode === 'full') {
        addendum = '/transactions/0';
    }
    return nock('https://bitgo.fakeurl')
        .persist(true)
        .post(`/api/v2/wallet/${params.walletId}/txrequests/${params.txRequestId}${addendum}/commit`)
        .reply(200, params.response);
}
function getRoute(tssType = 'eddsa', requestType = sdk_core_1.RequestType.tx, apiMode = 'lite') {
    switch (requestType) {
        case sdk_core_1.RequestType.tx:
            if (tssType === 'ecdsa' || apiMode === 'full') {
                return '/transactions/0';
            }
            break;
        case sdk_core_1.RequestType.message:
            if (tssType === 'ecdsa' || apiMode === 'full') {
                return '/messages/0';
            }
            break;
    }
    return '';
}
