// @bitgo/sdk-core/dist/src/bitgo/tss/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonTssMethods = exports.encryptYShare = exports.sendSignatureShare = exports.getTxRequest = exports.sendUserToBitgoGShare = exports.getBitgoToUserRShare = exports.offerUserToBitgoRShare = exports.createUserToBitGoGShare = exports.createUserSignShare = exports.createCombinedKey = exports.BitgoMpcGpgPubKeys = exports.ShareKeyPosition = exports.DKLSMethods = exports.ECDSAMethodTypes = exports.ECDSAMethods = exports.EDDSAMethodTypes = exports.EDDSAMethods = void 0;
const eddsa_1 = __importStar(require("./eddsa"));
exports.EDDSAMethods = eddsa_1.default;
Object.defineProperty(exports, "EDDSAMethodTypes", { enumerable: true, get: function () { return eddsa_1.EDDSAMethodTypes; } });
const ecdsa_1 = __importStar(require("./ecdsa"));
exports.ECDSAMethods = ecdsa_1.default;
Object.defineProperty(exports, "ECDSAMethodTypes", { enumerable: true, get: function () { return ecdsa_1.ECDSAMethodTypes; } });
Object.defineProperty(exports, "DKLSMethods", { enumerable: true, get: function () { return ecdsa_1.DKLSMethods; } });
var types_1 = require("./types");
Object.defineProperty(exports, "ShareKeyPosition", { enumerable: true, get: function () { return types_1.ShareKeyPosition; } });
exports.BitgoMpcGpgPubKeys = __importStar(require("./bitgoPubKeys"));
// exporting this types for backward compatibility.
/** @deprecated Use EDDSAMethods */
var eddsa_2 = require("./eddsa");
Object.defineProperty(exports, "createCombinedKey", { enumerable: true, get: function () { return eddsa_2.createCombinedKey; } });
Object.defineProperty(exports, "createUserSignShare", { enumerable: true, get: function () { return eddsa_2.createUserSignShare; } });
Object.defineProperty(exports, "createUserToBitGoGShare", { enumerable: true, get: function () { return eddsa_2.createUserToBitGoGShare; } });
Object.defineProperty(exports, "offerUserToBitgoRShare", { enumerable: true, get: function () { return eddsa_2.offerUserToBitgoRShare; } });
Object.defineProperty(exports, "getBitgoToUserRShare", { enumerable: true, get: function () { return eddsa_2.getBitgoToUserRShare; } });
Object.defineProperty(exports, "sendUserToBitgoGShare", { enumerable: true, get: function () { return eddsa_2.sendUserToBitgoGShare; } });
Object.defineProperty(exports, "getTxRequest", { enumerable: true, get: function () { return eddsa_2.getTxRequest; } });
Object.defineProperty(exports, "sendSignatureShare", { enumerable: true, get: function () { return eddsa_2.sendSignatureShare; } });
Object.defineProperty(exports, "encryptYShare", { enumerable: true, get: function () { return eddsa_2.encryptYShare; } });
exports.commonTssMethods = __importStar(require("./common"));
