// @bitgo/sdk-coin-trx/dist/test/unit/trx.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_assert_1 = __importDefault(require("node:assert"));
const node_test_1 = require("node:test");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const _ = __importStar(require("lodash"));
const src_1 = require("../../src");
const fixtures_1 = require("../fixtures");
const resources_1 = require("../resources");
(0, node_test_1.describe)('TRON:', function () {
    const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
    bitgo.initializeTestVars();
    bitgo.safeRegister('trx', src_1.Trx.createInstance);
    bitgo.safeRegister('ttrx', src_1.Ttrx.createInstance);
    let basecoin;
    // Test data from wallet 693b011a3ec26986f569b02140c7627e
    const testWalletData = {
        rootAddress: 'TAf36b36eqoMCzJJm3jwSsP81UvkMxrPbi',
        receiveAddress: 'TFaD6DeKFMcBuGuDD7LbbqxTnKunhXfdya',
        receiveAddressIndex: 2,
        keychains: [
            {
                id: '693b0110271fc3f5749754097793bb8d',
                pub: 'xpub661MyMwAqRbcFsVAdZyN2m8p21WHXg8NRNkqKApyS5gwmFsPdRTrmHYCnzR9vYe8DQ4uWGCBcAAsWE3r97HsFS3K2faZ2ejXNhHxdEoAEWC',
            },
            {
                id: '693b011065b9c4674825ce1f849b7bef',
                pub: 'xpub661MyMwAqRbcErKpZr9ztTFJk6fzXWatMFgRnpXfsjybWBfE9847EVGrHHBsGP8fcnzJmJuevAbPUEpjHTjEEWfYUWNMEDahvssQein848o',
            },
            {
                id: '693b01117c41846abb04818815b89b6c',
                pub: 'xpub661MyMwAqRbcFqZd7XU9DFW4f29VJzQt7UCA51ypaWa5ymhQ2pRZDTgViw3vZ56PqZ8dj1cracN3fAWhaiG1QKj9mvyt9Cba4nM2tPibNKw',
            },
        ],
    };
    (0, node_test_1.before)(function () {
        basecoin = bitgo.coin('ttrx');
    });
    (0, node_test_1.it)('should instantiate the coin', function () {
        const basecoin = bitgo.coin('trx');
        node_assert_1.default.ok(basecoin instanceof src_1.Trx);
    });
    (0, node_test_1.it)('explain a txHex', async function () {
        const txHex = JSON.stringify(fixtures_1.mockTx);
        const explainParams = {
            txHex,
            feeInfo: { fee: 1 },
            txID: fixtures_1.mockTx.txID,
        };
        const explanation = await basecoin.explainTransaction(explainParams);
        const toAddress = src_1.Utils.getBase58AddressFromHex(fixtures_1.mockTx.raw_data.contract[0].parameter.value.to_address);
        node_assert_1.default.equal(explanation.id, fixtures_1.mockTx.txID);
        node_assert_1.default.equal(explanation.outputs.length, 1);
        node_assert_1.default.equal(explanation.outputs[0].amount, '10');
        node_assert_1.default.equal(explanation.outputs[0].address, toAddress);
        node_assert_1.default.equal(explanation.outputAmount, '10');
        node_assert_1.default.equal(explanation.changeAmount, '0');
        node_assert_1.default.equal(explanation.changeOutputs.length, 0);
        node_assert_1.default.equal(explanation.fee.fee, 1);
        node_assert_1.default.equal(explanation.expiration, fixtures_1.mockTx.raw_data.expiration);
        node_assert_1.default.equal(explanation.timestamp, fixtures_1.mockTx.raw_data.timestamp);
    });
    (0, node_test_1.it)('should check valid addresses', function () {
        const badAddresses = [
            '',
            null,
            'xxxx',
            'YZ09fd-',
            '412C2BA4A9FF6C53207DC5B686BFECF75EA7B805772',
            '412C2BA4A9FF6C53207DC5B686BFECF75EA7B80',
            'TBChwKYNaTo4a4N68Me1qEiiKsRDspXqLLZ',
            '0x96be113992bdc3be24c11f6017085b605d253649',
            '0x341qg3922b1',
            '41E0C0F581D7D02D40826C1C6CBEE71F625D6344D0',
            '412C2BA4A9FF6C53207DC5B686BFECF75EA7B80577',
            '418840E6C55B9ADA326D211D818C34A994AECED808',
            '412A2B9F7641D0750C1E822D0E49EF765C8106524B',
            '41A614F803B6FD780986A42C78EC9C7F77E6DED13C',
            '418840E6C55B9ADA326D211D818C34A994AECED808',
        ];
        const goodAddresses = ['TBChwKYNaTo4a4N68Me1qEiiKsRDspXqLp', 'TPcf5jtYUhCN1X14tN577zF4NepbDZbxT7'];
        badAddresses.map((addr) => {
            node_assert_1.default.equal(basecoin.isValidAddress(addr), false);
        });
        goodAddresses.map((addr) => {
            node_assert_1.default.equal(basecoin.isValidAddress(addr), true);
        });
    });
    (0, node_test_1.it)('should throw if the params object is missing parameters', async function () {
        const explainParams = {
            feeInfo: { fee: 1 },
            txID: fixtures_1.mockTx.txID,
            txHex: null,
        };
        await node_assert_1.default.rejects(basecoin.explainTransaction(explainParams), {
            message: 'missing explain tx parameters',
        });
    });
    (0, node_test_1.it)('explain an half-signed/fully signed transaction', async function () {
        const txHex = JSON.stringify(fixtures_1.mockTx);
        const explainParams = {
            halfSigned: { txHex },
            feeInfo: { fee: 1 },
            txID: fixtures_1.mockTx.txID,
        };
        const explanation = await basecoin.explainTransaction(explainParams);
        const toAddress = src_1.Utils.getBase58AddressFromHex(fixtures_1.mockTx.raw_data.contract[0].parameter.value.to_address);
        node_assert_1.default.equal(explanation.id, fixtures_1.mockTx.txID);
        node_assert_1.default.equal(explanation.outputs.length, 1);
        node_assert_1.default.equal(explanation.outputs[0].amount, '10');
        node_assert_1.default.equal(explanation.outputs[0].address, toAddress);
        node_assert_1.default.equal(explanation.outputAmount, '10');
        node_assert_1.default.equal(explanation.changeAmount, '0');
        node_assert_1.default.equal(explanation.changeOutputs.length, 0);
        node_assert_1.default.equal(explanation.fee.fee, 1);
        node_assert_1.default.equal(explanation.expiration, fixtures_1.mockTx.raw_data.expiration);
        node_assert_1.default.equal(explanation.timestamp, fixtures_1.mockTx.raw_data.timestamp);
    });
    (0, node_test_1.it)('should sign a half signed tx', async function () {
        const tx = await basecoin.signTransaction(fixtures_1.signTxOptions);
        const unsignedTxJson = JSON.parse(fixtures_1.signTxOptions.txPrebuild.txHex);
        const signedTxJson = JSON.parse(tx.halfSigned.txHex);
        node_assert_1.default.equal(signedTxJson.txID, unsignedTxJson.txID);
        node_assert_1.default.equal(signedTxJson.raw_data_hex, unsignedTxJson.raw_data_hex);
        node_assert_1.default.deepStrictEqual(JSON.stringify(signedTxJson.raw_data), JSON.stringify(unsignedTxJson.raw_data));
        node_assert_1.default.equal(signedTxJson.signature.length, 1);
        node_assert_1.default.equal(signedTxJson.signature[0], '0a9944316924ec7fba4895f1ea1e7cc95f9e2b828ae268a48a8dbeddef40c6f5e127170a95aed9f3f5425b13058d0cb6ef1f5c2213190e482e87043691f22e6800');
    });
    (0, node_test_1.it)('should sign with an Xprv a half signed tx', async function () {
        const p = {
            prv: 'xprv9s21ZrQH143K2sg2Cukk5XqLQdrYnMCDah3y1FFVy6Hz9bQfqMSfmUiHPVHKhcUyft3N1emE5FudJVxgFm5N12MAg5o7DTPsDATTkwNgr73',
            txPrebuild: {
                txHex: fixtures_1.signTxOptions.txPrebuild.txHex,
            },
        };
        const tx = await basecoin.signTransaction(p);
        const unsignedTxJson = JSON.parse(fixtures_1.signTxOptions.txPrebuild.txHex);
        const signedTxJson = JSON.parse(tx.halfSigned.txHex);
        node_assert_1.default.equal(signedTxJson.txID, unsignedTxJson.txID);
        node_assert_1.default.equal(signedTxJson.raw_data_hex, unsignedTxJson.raw_data_hex);
        node_assert_1.default.deepStrictEqual(JSON.stringify(signedTxJson.raw_data), JSON.stringify(unsignedTxJson.raw_data));
        node_assert_1.default.equal(signedTxJson.signature.length, 1);
        node_assert_1.default.equal(signedTxJson.signature[0], '65e56f53a458c6f82d1ef39b2cf5be685a906ad22bb02699f907fcb72ef26f1e91cfc2b6a43bf5432faa0b63bdc5aebf1dc2f49a675d28d23fd7e038b3358b0600');
    });
    (0, node_test_1.it)('should add feeLimit to recipient', async function () {
        const feeLimit = 100;
        const buildParams = {
            recipients: [{ data: 'test' }],
            feeLimit,
        };
        basecoin.getExtraPrebuildParams(buildParams);
        node_assert_1.default.equal(buildParams.recipients[0].feeLimit, feeLimit);
    });
    (0, node_test_1.it)('should`t add any new field', async function () {
        const buildParams = {
            recipients: [{ data: 'test' }],
        };
        const unmodifiedBuildParams = _.cloneDeep(buildParams);
        await basecoin.getExtraPrebuildParams(buildParams);
        node_assert_1.default.deepStrictEqual(buildParams, unmodifiedBuildParams);
    });
    (0, node_test_1.describe)('Keypairs:', () => {
        (0, node_test_1.it)('should generate a keypair from random seed', function () {
            const keyPair = basecoin.generateKeyPair();
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(keyPair, 'pub'));
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(keyPair, 'prv'));
            node_assert_1.default.equal(basecoin.isValidPub(keyPair.pub), true);
        });
        (0, node_test_1.it)('should generate a keypair from a seed', function () {
            const seedText = '80350b4208d381fbfe2276a326603049fe500731c46d3c9936b5ce036b51377f24bab7dd0c2af7f107416ef858ff79b0670c72406dad064e72bb17fc0a9038bb';
            const seed = Buffer.from(seedText, 'hex');
            const keyPair = basecoin.generateKeyPair(seed);
            node_assert_1.default.equal(keyPair.pub, 'xpub661MyMwAqRbcFAwqvSGbk35kJf7CQqdN1w4CMUBBTqH5e3ivjU6D8ugv9hRSgRbRenC4w3ahXdLVahwjgjXhSuQKMdNdn55Y9TNSagBktws');
            node_assert_1.default.equal(keyPair.prv, 'xprv9s21ZrQH143K2gsNpQjbNu91kdGi1NuWei8bZ5mZuVk6mFPnBvmxb7NSJQdbZW3FGpK3Ycn7jorAXcEzMvviGtbyBz5tBrjfnWyQp3g75FK');
        });
    });
    (0, node_test_1.describe)('Build Unsigned Sweep', () => {
        (0, node_test_1.afterEach)(() => {
            node_test_1.mock.reset();
        });
        (0, node_test_1.it)('should recover trx from base address to recovery address', async function () {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.baseAddress) {
                    return Promise.resolve((0, resources_1.baseAddressBalance)(3000000));
                }
                return undefined;
            });
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            const destinationHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.recoveryDestination);
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === destinationHex && args[1] === baseAddrHex && args[2] === 900000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(baseAddrHex, destinationHex, 900000));
                }
                return undefined;
            });
            const res = await basecoin.recover({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                recoveryDestination: resources_1.TestRecoverData.recoveryDestination,
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'txHex'));
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'feeInfo'));
            const rawData = JSON.parse(res.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData, 'contract'));
            const value = rawData.contract[0].parameter.value;
            node_assert_1.default.equal(value.amount, 900000);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.owner_address), resources_1.TestRecoverData.baseAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.to_address), resources_1.TestRecoverData.recoveryDestination);
        });
        (0, node_test_1.it)('should recover trx from base address to recovery address with scan passed as valid integer string', async function () {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.baseAddress) {
                    return Promise.resolve((0, resources_1.baseAddressBalance)(3000000));
                }
                return undefined;
            });
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            const destinationHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.recoveryDestination);
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === destinationHex && args[1] === baseAddrHex && args[2] === 900000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(baseAddrHex, destinationHex, 900000));
                }
                return undefined;
            });
            const res = await basecoin.recover({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                recoveryDestination: resources_1.TestRecoverData.recoveryDestination,
                scan: '10',
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'txHex'));
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'feeInfo'));
            const rawData = JSON.parse(res.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData, 'contract'));
            const value = rawData.contract[0].parameter.value;
            node_assert_1.default.equal(value.amount, 900000);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.owner_address), resources_1.TestRecoverData.baseAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.to_address), resources_1.TestRecoverData.recoveryDestination);
        });
        (0, node_test_1.it)('should fail recover trx from base address to recovery address with scan passed as valid integer string', async function () {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.baseAddress) {
                    return Promise.resolve((0, resources_1.baseAddressBalance)(3000000));
                }
                return undefined;
            });
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            const destinationHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.recoveryDestination);
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === destinationHex && args[1] === baseAddrHex && args[2] === 900000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(baseAddrHex, destinationHex, 900000));
                }
                return undefined;
            });
            await node_assert_1.default.rejects(basecoin.recover({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                recoveryDestination: resources_1.TestRecoverData.recoveryDestination,
                scan: 'abc',
            }), {
                message: 'Invalid scanning factor',
            });
        });
        (0, node_test_1.it)('should recover trx from receive address to base address', async function () {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.baseAddress) {
                    return Promise.resolve((0, resources_1.baseAddressBalance)(2000000));
                }
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.firstReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(102100000, resources_1.TestRecoverData.firstReceiveAddress));
                }
                return undefined;
            });
            const firstReceiveAddressHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.firstReceiveAddress);
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === baseAddrHex && args[1] === firstReceiveAddressHex && args[2] === 100000000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(firstReceiveAddressHex, baseAddrHex, 100000000));
                }
                return undefined;
            });
            const res = await basecoin.recover({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                recoveryDestination: resources_1.TestRecoverData.recoveryDestination,
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'txHex'));
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'feeInfo'));
            const rawData = JSON.parse(res.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData, 'contract'));
            const value = rawData.contract[0].parameter.value;
            node_assert_1.default.equal(value.amount, 100000000);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.owner_address), resources_1.TestRecoverData.firstReceiveAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.to_address), resources_1.TestRecoverData.baseAddress);
        });
        (0, node_test_1.it)('should recover token from base address to recovery address', async function () {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.baseAddress) {
                    return Promise.resolve((0, resources_1.baseAddressBalance)(100000000, [
                        {
                            TSdZwNqpHofzP6BsBKGQUWdBeJphLmF6id: '1000000000',
                        },
                        {
                            TG3XXyExBkPp9nzdajDZsozEu4BkaSJozs: '1100000000',
                        },
                    ]));
                }
                return undefined;
            });
            node_test_1.mock.method(src_1.Trx.prototype, 'getTriggerSmartContractTransaction', (...args) => {
                return Promise.resolve(resources_1.SampleRawTokenSendTxn);
            });
            const res = await basecoin.recover({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                tokenContractAddress: 'TG3XXyExBkPp9nzdajDZsozEu4BkaSJozs',
                recoveryDestination: resources_1.TestRecoverData.recoveryDestination,
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.equal(res.recoveryAmount, 1100000000);
            node_assert_1.default.equal(res.feeInfo.fee, '100000000');
            const expirationDuration = res.tx.raw_data.expiration - res.tx.raw_data.timestamp;
            node_assert_1.default.ok(expirationDuration >= 86400000);
            node_assert_1.default.equal(res.addressInfo, undefined);
            const rawData = JSON.parse(res.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData, 'contract'));
            const value = rawData.contract[0].parameter.value;
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.owner_address), resources_1.TestRecoverData.baseAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.contract_address), 'TG3XXyExBkPp9nzdajDZsozEu4BkaSJozs');
        });
        (0, node_test_1.it)('should throw if trx balance at base address is not sufficient to cover token send', async function () {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.baseAddress) {
                    return Promise.resolve((0, resources_1.baseAddressBalance)(1000000, [
                        {
                            TG3XXyExBkPp9nzdajDZsozEu4BkaSJozs: '1100000000',
                        },
                    ]));
                }
                return undefined;
            });
            node_test_1.mock.method(src_1.Trx.prototype, 'getTriggerSmartContractTransaction', (...args) => {
                return Promise.resolve(resources_1.SampleRawTokenSendTxn);
            });
            await node_assert_1.default.rejects(basecoin.recover({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                tokenContractAddress: 'TG3XXyExBkPp9nzdajDZsozEu4BkaSJozs',
                recoveryDestination: resources_1.TestRecoverData.recoveryDestination,
            }), {
                message: "Amount of funds to recover 1000000 is less than 100000000 and wouldn't be able to fund a trc20 send",
            });
        });
    });
    (0, node_test_1.describe)('Build Unsigned Consolidation Recoveries', () => {
        (0, node_test_1.afterEach)(() => {
            node_test_1.mock.reset();
        });
        (0, node_test_1.it)('should throw if startingScanIndex is not ge to 1', async () => {
            await node_assert_1.default.rejects(basecoin.recoverConsolidations({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                startingScanIndex: -1,
            }), {
                message: 'Invalid starting or ending index to scan for addresses. startingScanIndex: -1, endingScanIndex: 19.',
            });
        });
        (0, node_test_1.it)('should throw if scan factor is too high', async () => {
            await node_assert_1.default.rejects(basecoin.recoverConsolidations({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                startingScanIndex: 1,
                endingScanIndex: 300,
            }), {
                message: 'Invalid starting or ending index to scan for addresses. startingScanIndex: 1, endingScanIndex: 300.',
            });
        });
        (0, node_test_1.it)('should build consolidate recoveries', async () => {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.firstReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(102100000, resources_1.TestRecoverData.firstReceiveAddress));
                }
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.secondReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(50000000, resources_1.TestRecoverData.secondReceiveAddress));
                }
                return undefined;
            });
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === baseAddrHex && args[1] === firstReceiveAddrHex && args[2] === 100000000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(firstReceiveAddrHex, baseAddrHex, 100000000));
                }
                if (args.length > 0 && args[0] === baseAddrHex && args[1] === secondReceiveAddrHex && args[2] === 47900000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(secondReceiveAddrHex, baseAddrHex, 47900000));
                }
                return undefined;
            });
            const firstReceiveAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.firstReceiveAddress);
            const secondReceiveAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.secondReceiveAddress);
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            const res = await basecoin.recoverConsolidations({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                startingScanIndex: 1,
                endingScanIndex: 3,
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'transactions'));
            node_assert_1.default.equal(res.transactions.length, 2);
            const txn1 = res.transactions[0];
            const rawData1 = JSON.parse(txn1.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData1, 'contract'));
            const value1 = rawData1.contract[0].parameter.value;
            node_assert_1.default.equal(value1.amount, 100000000);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value1.owner_address), resources_1.TestRecoverData.firstReceiveAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value1.to_address), resources_1.TestRecoverData.baseAddress);
            const txn2 = res.transactions[1];
            const rawData2 = JSON.parse(txn2.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData2, 'contract'));
            const value2 = rawData2.contract[0].parameter.value;
            node_assert_1.default.equal(value2.amount, 47900000);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value2.owner_address), resources_1.TestRecoverData.secondReceiveAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value2.to_address), resources_1.TestRecoverData.baseAddress);
        });
        (0, node_test_1.it)('should build consolidate token recoveries', async () => {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.firstReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(202100000, resources_1.TestRecoverData.firstReceiveAddress, [
                        {
                            TSdZwNqpHofzP6BsBKGQUWdBeJphLmF6id: '1100000000',
                        },
                    ]));
                }
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.secondReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(500, resources_1.TestRecoverData.secondReceiveAddress));
                }
                return undefined;
            });
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === baseAddrHex && args[1] === firstReceiveAddrHex && args[2] === 1100000000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(firstReceiveAddrHex, baseAddrHex, 1100000000));
                }
                return undefined;
            });
            const firstReceiveAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.firstReceiveAddress);
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            const res = await basecoin.recoverConsolidations({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                tokenContractAddress: 'TSdZwNqpHofzP6BsBKGQUWdBeJphLmF6id',
                startingScanIndex: 1,
                endingScanIndex: 3,
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'transactions'));
            node_assert_1.default.equal(res.transactions.length, 1);
            const txn = res.transactions[0];
            const rawData = JSON.parse(txn.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData, 'contract'));
            const value = rawData.contract[0].parameter.value;
            node_assert_1.default.equal(value.data, 'a9059cbb000000000000000000000000c25420255c2c5a2dd54ef69f92ef261e6bd4216a000000000000000000000000000000000000000000000000000000004190ab00');
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.owner_address), resources_1.TestRecoverData.firstReceiveAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value.contract_address), 'TSdZwNqpHofzP6BsBKGQUWdBeJphLmF6id');
        });
        (0, node_test_1.it)('should skip building consolidate transaction if balance is lower than reserved fee', async () => {
            node_test_1.mock.method(src_1.Trx.prototype, 'getAccountBalancesFromNode', (...args) => {
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.firstReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(102100000, resources_1.TestRecoverData.firstReceiveAddress));
                }
                if (args.length > 0 && args[0] === resources_1.TestRecoverData.secondReceiveAddress) {
                    return Promise.resolve((0, resources_1.receiveAddressBalance)(2000000, resources_1.TestRecoverData.secondReceiveAddress));
                }
                return undefined;
            });
            node_test_1.mock.method(src_1.Trx.prototype, 'getBuildTransaction', (...args) => {
                if (args.length > 0 && args[0] === baseAddrHex && args[1] === firstReceiveAddrHex && args[2] === 100000000) {
                    return Promise.resolve((0, resources_1.creationTransaction)(firstReceiveAddrHex, baseAddrHex, 100000000));
                }
                return undefined;
            });
            const firstReceiveAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.firstReceiveAddress);
            const baseAddrHex = src_1.Utils.getHexAddressFromBase58Address(resources_1.TestRecoverData.baseAddress);
            const res = await basecoin.recoverConsolidations({
                userKey: resources_1.TestRecoverData.userKey,
                backupKey: resources_1.TestRecoverData.backupKey,
                bitgoKey: resources_1.TestRecoverData.bitgoKey,
                startingScanIndex: 1,
                endingScanIndex: 3,
            });
            node_assert_1.default.notEqual(res.length, 0);
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(res, 'transactions'));
            node_assert_1.default.equal(res.transactions.length, 1);
            const txn1 = res.transactions[0];
            const rawData1 = JSON.parse(txn1.txHex).raw_data;
            node_assert_1.default.ok(Object.prototype.hasOwnProperty.call(rawData1, 'contract'));
            const value1 = rawData1.contract[0].parameter.value;
            node_assert_1.default.equal(value1.amount, 100000000);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value1.owner_address), resources_1.TestRecoverData.firstReceiveAddress);
            node_assert_1.default.equal(src_1.Utils.getBase58AddressFromHex(value1.to_address), resources_1.TestRecoverData.baseAddress);
        });
    });
    (0, node_test_1.describe)('isWalletAddress', () => {
        (0, node_test_1.it)('should verify root address (index 0)', async function () {
            const result = await basecoin.isWalletAddress({
                address: testWalletData.rootAddress,
                keychains: testWalletData.keychains,
                index: 0,
            });
            node_assert_1.default.equal(result, true);
        });
        (0, node_test_1.it)('should verify receive address (index > 0)', async function () {
            const result = await basecoin.isWalletAddress({
                address: testWalletData.receiveAddress,
                keychains: testWalletData.keychains,
                index: testWalletData.receiveAddressIndex,
                chain: 0,
            });
            node_assert_1.default.equal(result, true);
        });
        (0, node_test_1.it)('should verify address with index in coinSpecific', async function () {
            const result = await basecoin.isWalletAddress({
                address: testWalletData.rootAddress,
                keychains: testWalletData.keychains,
                coinSpecific: { index: 0 },
            });
            node_assert_1.default.equal(result, true);
        });
        (0, node_test_1.it)('should verify address with string index', async function () {
            const result = await basecoin.isWalletAddress({
                address: testWalletData.rootAddress,
                keychains: testWalletData.keychains,
                index: '0',
            });
            node_assert_1.default.equal(result, true);
        });
        (0, node_test_1.it)('should return false for wrong address at index 0', async function () {
            const result = await basecoin.isWalletAddress({
                address: testWalletData.receiveAddress, // wrong address for index 0
                keychains: testWalletData.keychains,
                index: 0,
            });
            node_assert_1.default.equal(result, false);
        });
        (0, node_test_1.it)('should return false for wrong address at receive index', async function () {
            const result = await basecoin.isWalletAddress({
                address: testWalletData.rootAddress, // wrong address for index 1
                keychains: testWalletData.keychains,
                index: testWalletData.receiveAddressIndex,
                chain: 0,
            });
            node_assert_1.default.equal(result, false);
        });
        (0, node_test_1.it)('should throw for invalid address', async function () {
            await node_assert_1.default.rejects(basecoin.isWalletAddress({
                address: 'invalid-address',
                keychains: testWalletData.keychains,
                index: 0,
            }), {
                message: 'Invalid address: invalid-address',
            });
        });
        (0, node_test_1.it)('should throw for missing index', async function () {
            await node_assert_1.default.rejects(basecoin.isWalletAddress({
                address: testWalletData.rootAddress,
                keychains: testWalletData.keychains,
            }), {
                message: 'Invalid or missing index for address verification',
            });
        });
        (0, node_test_1.it)('should throw for missing bitgo key on root address verification', async function () {
            await node_assert_1.default.rejects(basecoin.isWalletAddress({
                address: testWalletData.rootAddress,
                keychains: testWalletData.keychains.slice(0, 2), // only user and backup keys
                index: 0,
            }), {
                message: 'BitGo public key required for root address verification',
            });
        });
        (0, node_test_1.it)('should throw for missing user key on receive address verification', async function () {
            await node_assert_1.default.rejects(basecoin.isWalletAddress({
                address: testWalletData.receiveAddress,
                keychains: [], // no keys
                index: testWalletData.receiveAddressIndex,
            }), {
                message: 'User public key required for receive address verification',
            });
        });
    });
});
