// @bitgo/sdk-coin-trx/dist/test/unit/verifyTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_assert_1 = __importDefault(require("node:assert"));
const crypto_1 = require("crypto");
const node_test_1 = require("node:test");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../src");
const lib_1 = require("../../src/lib");
const resources_1 = require("../resources");
(0, node_test_1.describe)('TRON Verify Transaction:', function () {
    const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
    bitgo.initializeTestVars();
    bitgo.safeRegister('trx', src_1.Trx.createInstance);
    bitgo.safeRegister('ttrx', src_1.Ttrx.createInstance);
    let basecoin;
    (0, node_test_1.before)(function () {
        basecoin = bitgo.coin('ttrx');
    });
    (0, node_test_1.describe)('Parameter Validation', () => {
        (0, node_test_1.it)('should throw error when txParams is missing', async function () {
            const params = {
                txPrebuild: {
                    txHex: JSON.stringify({
                        ...resources_1.UnsignedBuildTransaction,
                        raw_data: {
                            ...resources_1.UnsignedBuildTransaction.raw_data,
                            expiration: Date.now() + 3600000,
                            timestamp: Date.now(),
                        },
                    }),
                },
                wallet: {},
            };
            await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                message: 'missing txParams',
            });
        });
        (0, node_test_1.it)('should throw error when wallet or txPrebuild is missing', async function () {
            const params = {
                txParams: {
                    recipients: [{ address: 'TQFxDSoXy2yXRE5HtKwAwrNRXGxYxkeSGk', amount: '1000000' }],
                },
            };
            await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                message: 'missing txPrebuild',
            });
        });
        (0, node_test_1.it)('should throw error when txPrebuild.txHex is missing', async function () {
            const params = {
                txParams: {
                    recipients: [{ address: 'TQFxDSoXy2yXRE5HtKwAwrNRXGxYxkeSGk', amount: '1000000' }],
                },
                txPrebuild: {},
                wallet: {},
            };
            await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                message: 'missing txHex in txPrebuild',
            });
        });
    });
    (0, node_test_1.describe)('Contract Type Validation', () => {
        (0, node_test_1.describe)('TransferContract', () => {
            (0, node_test_1.it)('should validate valid TransferContract', async function () {
                const timestamp = Date.now();
                const transferContract = {
                    parameter: {
                        value: {
                            amount: 1000000,
                            owner_address: '4173a5993cd182ae152adad8203163f780c65a8aa5',
                            to_address: '41d6cd6a2c0ff35a319e6abb5b9503ba0278679882',
                        },
                        type_url: 'type.googleapis.com/protocol.TransferContract',
                    },
                    type: 'TransferContract',
                };
                const rawData = {
                    contract: [transferContract],
                    ref_block_bytes: 'c8cf',
                    ref_block_hash: '89177fd84c5d9196',
                    expiration: timestamp + 3600000,
                    timestamp: timestamp,
                    fee_limit: 150000000,
                };
                // Transform rawData to match the expected parameter structure
                const transformedRawData = {
                    contract: rawData.contract,
                    refBlockBytes: rawData.ref_block_bytes,
                    refBlockHash: rawData.ref_block_hash,
                    expiration: rawData.expiration,
                    timestamp: rawData.timestamp,
                    feeLimit: rawData.fee_limit,
                };
                // Generate raw_data_hex using the utility function
                const rawDataHex = lib_1.Utils.generateRawDataHex(transformedRawData);
                // Calculate txID as SHA256 hash of raw_data_hex
                const txID = (0, crypto_1.createHash)('sha256').update(Buffer.from(rawDataHex, 'hex')).digest('hex');
                const params = {
                    txParams: {
                        recipients: [
                            {
                                address: lib_1.Utils.getBase58AddressFromHex('41d6cd6a2c0ff35a319e6abb5b9503ba0278679882'),
                                amount: '1000000',
                            },
                        ],
                    },
                    txPrebuild: {
                        txHex: JSON.stringify({
                            txID,
                            raw_data: rawData,
                            raw_data_hex: rawDataHex,
                        }),
                    },
                    wallet: {},
                };
                const result = await basecoin.verifyTransaction(params);
                node_assert_1.default.strictEqual(result, true);
            });
            (0, node_test_1.it)('should fail with missing owner address', async function () {
                const timestamp = Date.now();
                const txID = '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef';
                const transferContract = {
                    parameter: {
                        value: {
                            amount: 1000000,
                            to_address: '41c25420255c2c5a2dd54ef69f92ef261e6bd4216a',
                        },
                        type_url: 'type.googleapis.com/protocol.TransferContract',
                    },
                    type: 'TransferContract',
                };
                const rawData = {
                    txID,
                    contract: [transferContract],
                    ref_block_bytes: 'c8cf',
                    ref_block_hash: '89177fd84c5d9196',
                    expiration: timestamp + 3600000,
                    timestamp: timestamp,
                    fee_limit: 150000000,
                };
                const transformedRawData = {
                    contract: rawData.contract,
                    refBlockBytes: rawData.ref_block_bytes,
                    refBlockHash: rawData.ref_block_hash,
                    expiration: rawData.expiration,
                    timestamp: rawData.timestamp,
                    feeLimit: rawData.fee_limit,
                };
                const expectedRawDataHex = lib_1.Utils.generateRawDataHex(transformedRawData);
                const params = {
                    txParams: {
                        recipients: [{ address: 'TLWh67P93KgtnZNCtGnEHM1H33Nhq2uvvN', amount: '1000000' }],
                    },
                    txPrebuild: {
                        txHex: JSON.stringify({
                            txID,
                            raw_data: rawData,
                            raw_data_hex: expectedRawDataHex,
                        }),
                    },
                    wallet: {},
                };
                await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                    message: 'Transaction has not have a valid id',
                });
            });
            (0, node_test_1.it)('should fail with missing destination address', async function () {
                const timestamp = Date.now();
                const txID = '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef';
                const transferContract = {
                    parameter: {
                        value: {
                            amount: 1000000,
                            owner_address: '4173a5993cd182ae152adad8203163f780c65a8aa5',
                        },
                        type_url: 'type.googleapis.com/protocol.TransferContract',
                    },
                    type: 'TransferContract',
                };
                const rawData = {
                    txID,
                    contract: [transferContract],
                    ref_block_bytes: 'c8cf',
                    ref_block_hash: '89177fd84c5d9196',
                    expiration: timestamp + 3600000,
                    timestamp: timestamp,
                    fee_limit: 150000000,
                };
                const transformedRawData = {
                    contract: rawData.contract,
                    refBlockBytes: rawData.ref_block_bytes,
                    refBlockHash: rawData.ref_block_hash,
                    expiration: rawData.expiration,
                    timestamp: rawData.timestamp,
                    feeLimit: rawData.fee_limit,
                };
                const expectedRawDataHex = lib_1.Utils.generateRawDataHex(transformedRawData);
                const params = {
                    txParams: {
                        recipients: [{ address: 'TLWh67P93KgtnZNCtGnEHM1H33Nhq2uvvN', amount: '1000000' }],
                    },
                    txPrebuild: {
                        txHex: JSON.stringify({
                            txID,
                            raw_data: rawData,
                            raw_data_hex: expectedRawDataHex,
                        }),
                    },
                    wallet: {},
                };
                await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                    message: 'Transaction has not have a valid id',
                });
            });
            (0, node_test_1.it)('should fail with missing amount', async function () {
                const timestamp = Date.now();
                const transferContract = {
                    parameter: {
                        value: {
                            owner_address: '4173a5993cd182ae152adad8203163f780c65a8aa5',
                            to_address: '41c25420255c2c5a2dd54ef69f92ef261e6bd4216a',
                        },
                        type_url: 'type.googleapis.com/protocol.TransferContract',
                    },
                    type: 'TransferContract',
                };
                const rawData = {
                    contract: [transferContract],
                    ref_block_bytes: 'c8cf',
                    ref_block_hash: '89177fd84c5d9196',
                    expiration: timestamp + 3600000,
                    timestamp: timestamp,
                    fee_limit: 150000000,
                };
                const transformedRawData = {
                    contract: rawData.contract,
                    refBlockBytes: rawData.ref_block_bytes,
                    refBlockHash: rawData.ref_block_hash,
                    expiration: rawData.expiration,
                    timestamp: rawData.timestamp,
                    feeLimit: rawData.fee_limit,
                };
                const expectedRawDataHex = lib_1.Utils.generateRawDataHex(transformedRawData);
                const params = {
                    txParams: {
                        recipients: [{ address: 'TLWh67P93KgtnZNCtGnEHM1H33Nhq2uvvN', amount: '1000000' }],
                    },
                    txPrebuild: {
                        txHex: JSON.stringify({
                            raw_data: rawData,
                            raw_data_hex: expectedRawDataHex,
                        }),
                    },
                    wallet: {},
                };
                await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                    message: 'Amount does not exist in this transfer contract.',
                });
            });
            (0, node_test_1.it)('should fail due to amount missmatch', async function () {
                const timestamp = Date.now();
                const transferContract = {
                    parameter: {
                        value: {
                            amount: 2000000,
                            owner_address: '4173a5993cd182ae152adad8203163f780c65a8aa5',
                            to_address: '41d6cd6a2c0ff35a319e6abb5b9503ba0278679882',
                        },
                        type_url: 'type.googleapis.com/protocol.TransferContract',
                    },
                    type: 'TransferContract',
                };
                const rawData = {
                    contract: [transferContract],
                    ref_block_bytes: 'c8cf',
                    ref_block_hash: '89177fd84c5d9196',
                    expiration: timestamp + 3600000,
                    timestamp: timestamp,
                    fee_limit: 150000000,
                };
                // Transform rawData to match the expected parameter structure
                const transformedRawData = {
                    contract: rawData.contract,
                    refBlockBytes: rawData.ref_block_bytes,
                    refBlockHash: rawData.ref_block_hash,
                    expiration: rawData.expiration,
                    timestamp: rawData.timestamp,
                    feeLimit: rawData.fee_limit,
                };
                // Generate raw_data_hex using the utility function
                const rawDataHex = lib_1.Utils.generateRawDataHex(transformedRawData);
                // Calculate txID as SHA256 hash of raw_data_hex
                const txID = (0, crypto_1.createHash)('sha256').update(Buffer.from(rawDataHex, 'hex')).digest('hex');
                const params = {
                    txParams: {
                        recipients: [
                            {
                                address: '41d6cd6a2c0ff35a319e6abb5b9503ba0278679882',
                                amount: '1000000',
                            },
                        ],
                    },
                    txPrebuild: {
                        txHex: JSON.stringify({
                            txID,
                            raw_data: rawData,
                            raw_data_hex: rawDataHex,
                        }),
                    },
                    wallet: {},
                };
                await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                    message: 'transaction amount in txPrebuild does not match the value given by client',
                });
            });
            (0, node_test_1.it)('should fail due to destination address missmatch', async function () {
                const timestamp = Date.now();
                const transferContract = {
                    parameter: {
                        value: {
                            amount: 1000000,
                            owner_address: '4173a5993cd182ae152adad8203163f780c65a8aa5',
                            to_address: '41d6cd6a2c0ff35a319e6abb5b9503ba0278679882',
                        },
                        type_url: 'type.googleapis.com/protocol.TransferContract',
                    },
                    type: 'TransferContract',
                };
                const rawData = {
                    contract: [transferContract],
                    ref_block_bytes: 'c8cf',
                    ref_block_hash: '89177fd84c5d9196',
                    expiration: timestamp + 3600000,
                    timestamp: timestamp,
                    fee_limit: 150000000,
                };
                // Transform rawData to match the expected parameter structure
                const transformedRawData = {
                    contract: rawData.contract,
                    refBlockBytes: rawData.ref_block_bytes,
                    refBlockHash: rawData.ref_block_hash,
                    expiration: rawData.expiration,
                    timestamp: rawData.timestamp,
                    feeLimit: rawData.fee_limit,
                };
                // Generate raw_data_hex using the utility function
                const rawDataHex = lib_1.Utils.generateRawDataHex(transformedRawData);
                // Calculate txID as SHA256 hash of raw_data_hex
                const txID = (0, crypto_1.createHash)('sha256').update(Buffer.from(rawDataHex, 'hex')).digest('hex');
                const params = {
                    txParams: {
                        recipients: [
                            {
                                address: '41d6cd6a2c0ff35a319e6abb5b9503ba0278679883',
                                amount: '1000000',
                            },
                        ],
                    },
                    txPrebuild: {
                        txHex: JSON.stringify({
                            txID,
                            raw_data: rawData,
                            raw_data_hex: rawDataHex,
                        }),
                    },
                    wallet: {},
                };
                await node_assert_1.default.rejects(basecoin.verifyTransaction(params), {
                    message: 'destination address does not match with the recipient address',
                });
            });
        });
    });
});
