// @bitgo/sdk-coin-flr/dist/src/flrToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlrToken = void 0;
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class FlrToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, FlrToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, FlrToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(FlrToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc **/
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /**
     * Make a query to Flare explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @param {string} apiKey optional API key to use instead of the one from the environment
     * @returns {Promise<Object>} response from Flare explorer
     */
    async recoveryBlockchainExplorerQuery(query, apiKey) {
        const apiToken = apiKey || sdk_core_1.common.Environments[this.bitgo.getEnv()].flrExplorerApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].flrExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    getFullName() {
        return 'Flare Token';
    }
}
exports.FlrToken = FlrToken;
FlrToken.coinNames = {
    Mainnet: 'flr',
    Testnet: 'tflr',
};
