// @bitgo/account-lib/dist/test/unit/fixtures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockMessage = exports.MockMessageBuilder = exports.MockMessageBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
class MockMessageBuilderFactory extends sdk_core_1.BaseMessageBuilderFactory {
    constructor(coinConfig) {
        super(coinConfig);
    }
    getMessageBuilder(type) {
        return new MockMessageBuilder(this.coinConfig, type);
    }
}
exports.MockMessageBuilderFactory = MockMessageBuilderFactory;
class MockMessageBuilder extends sdk_core_1.BaseMessageBuilder {
    constructor(coinConfig, type = sdk_core_1.MessageStandardType.UNKNOWN) {
        super(coinConfig, type);
    }
    async buildMessage(options) {
        return new MockMessage(options);
    }
    async fromBroadcastFormat(broadcastMessage) {
        this.setType(broadcastMessage.type);
        this.setPayload(broadcastMessage.payload);
        this.setSignatures(broadcastMessage.signatures || []);
        this.setSigners(broadcastMessage.signers || []);
        if (broadcastMessage.metadata) {
            this.setMetadata(broadcastMessage.metadata);
        }
        return this.build();
    }
}
exports.MockMessageBuilder = MockMessageBuilder;
class MockMessage extends sdk_core_1.BaseMessage {
    constructor(options) {
        super(options);
    }
    async getSignablePayload() {
        this.signablePayload = Buffer.from(this.payload);
        return this.signablePayload;
    }
}
exports.MockMessage = MockMessage;
