// @bitgo/sdk-coin-stx/dist/src/sip10Token.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sip10Token = void 0;
const lodash_1 = __importDefault(require("lodash"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const statics_1 = require("@bitgo/statics");
const stx_1 = require("./stx");
const lib_1 = require("./lib");
const utils_1 = require("./lib/utils");
class Sip10Token extends stx_1.Stx {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === statics_1.NetworkType.MAINNET ? statics_1.coins.get('stx') : statics_1.coins.get('tstx');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new Sip10Token(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.stx.tokens, ...statics_1.tokens.testnet.stx.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            const tokenConstructor = Sip10Token.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get assetId() {
        return this.tokenConfig.assetId;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Sip10 Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    getTransaction(coinConfig) {
        return new lib_1.TransactionBuilderFactory(coinConfig).getFungibleTokenTransferBuilder();
    }
    async verifyTransaction(params) {
        const { txPrebuild: txPrebuild, txParams: txParams } = params;
        const { memo } = txParams;
        if (Array.isArray(txParams.recipients) && txParams.recipients.length > 1) {
            throw new Error(`${this.getChain()} doesn't support sending to more than 1 destination address within a single transaction. Try again, using only a single recipient.`);
        }
        const rawTx = txPrebuild.txHex;
        if (!rawTx) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const coinConfig = statics_1.coins.get(this.getChain());
        const transaction = this.getTransaction(coinConfig);
        transaction.from(rawTx);
        const explainedTx = await this.explainTransaction({ txHex: rawTx, feeInfo: { fee: '' } });
        if (txParams.recipients !== undefined && explainedTx) {
            const filteredRecipients = txParams.recipients?.map((recipient) => {
                const addressDetails = (0, utils_1.getMemoIdAndBaseAddressFromAddress)(recipient.address);
                const recipientData = {
                    address: addressDetails.address,
                    amount: BigInt(recipient.amount),
                };
                if (recipient.tokenName) {
                    recipientData['tokenName'] = recipient.tokenName;
                }
                return recipientData;
            });
            const filteredOutputs = explainedTx.outputs.map((output) => {
                const recipientData = {
                    address: output.address,
                    amount: BigInt(output.amount),
                };
                if (output.tokenName) {
                    recipientData['tokenName'] = output.tokenName;
                }
                return recipientData;
            });
            if (!lodash_1.default.isEqual(filteredOutputs, filteredRecipients)) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            // compare memo
            let memoInput = '';
            let memoOutput = '';
            if (memo && memo.value) {
                memoInput = memo.value;
            }
            else if (txParams.recipients.length) {
                const addressDetails = (0, utils_1.getMemoIdAndBaseAddressFromAddress)(txParams.recipients[0].address);
                memoInput = addressDetails.memoId ? addressDetails.memoId : '';
            }
            if (explainedTx.memo) {
                memoOutput = explainedTx.memo;
            }
            if (!lodash_1.default.isEqual(memoInput, memoOutput)) {
                throw new Error('Tx memo does not match with expected txParams recipient memo');
            }
            // compare send amount
            let totalAmount = new bignumber_js_1.default(0);
            for (const recipients of txParams.recipients) {
                totalAmount = totalAmount.plus(recipients.amount);
            }
            if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                throw new Error('Tx total amount does not match with expected total amount field');
            }
        }
        return true;
    }
}
exports.Sip10Token = Sip10Token;
