// @bitgo/sdk-coin-cosmos/dist/test/testUtils/utils.js
"use strict";
/**
 * Utility functions for Cosmos SDK test data
 * This file contains utility functions for test data
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBuilderFactory = exports.getAllTestData = exports.getTestData = exports.getAvailableTestCoins = void 0;
exports.ensureTransaction = ensureTransaction;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const statics_1 = require("@bitgo/statics");
const transactionBuilderFactory_1 = require("../../src/lib/transactionBuilderFactory");
/**
 * Get all available test coins
 * @returns {string[]} Array of coin names that have test data
 */
const getAvailableTestCoins = () => {
    try {
        // Get the resources directory path
        const resourcesDir = path.join(__dirname, '../resources');
        // Read all files in the resources directory
        const files = fs.readdirSync(resourcesDir);
        // Extract coin names by removing the .ts extension
        return files.map((file) => file.replace('.ts', ''));
    }
    catch (error) {
        // Fallback to hardcoded list if there's an error
        console.warn('Failed to dynamically discover test coins:', error);
        return ['cosmos', 'cronos'];
    }
};
exports.getAvailableTestCoins = getAvailableTestCoins;
/**
 * Load test data for a specific coin
 *
 * This utility abstracts away the path resolution logic,
 * making it easier to load test data from anywhere in the test suite.
 *
 * @param {string} coinName - The coin name (e.g., 'cosmos', 'cronos')
 * @returns {CoinTestData} The test data for the coin
 * @throws {Error} If the test data for the coin is not found
 */
const getTestData = (coinName) => {
    try {
        // Dynamic import of the coin-specific test data
        return require(`../resources/${coinName}`).default;
    }
    catch (e) {
        throw new Error(`Test data for coin ${coinName} not found: ${e}`);
    }
};
exports.getTestData = getTestData;
/**
 * Get test data for all available coins
 *
 * @returns {Record<string, CoinTestData>} An object mapping coin names to their test data
 */
const getAllTestData = () => {
    const availableCoins = (0, exports.getAvailableTestCoins)();
    const result = {};
    for (const coin of availableCoins) {
        result[coin] = (0, exports.getTestData)(coin);
    }
    return result;
};
exports.getAllTestData = getAllTestData;
/**
 * Get the builder factory for a specific coin
 * @param {string} coin - The coin name
 * @returns {TransactionBuilderFactory} The transaction builder factory
 */
const getBuilderFactory = (coin) => {
    const coinConfig = statics_1.coins.get(coin);
    return new transactionBuilderFactory_1.TransactionBuilderFactory(coinConfig);
};
exports.getBuilderFactory = getBuilderFactory;
/**
 * Ensures that all required properties are present in the test transaction data
 * This is useful for TypeScript type checking in test files
 *
 * @param {T} tx - The transaction data that might have optional properties
 * @returns {Required<T>} The same transaction data but with TypeScript treating all properties as non-optional
 */
function ensureTransaction(tx) {
    // This function doesn't actually modify the data
    // It just tells TypeScript that all properties are present
    return tx;
}
