// @bitgo/sdk-coin-injective/dist/test/unit/injective.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const sdk_test_1 = require("@bitgo/sdk-test");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const mocha_1 = require("mocha");
const sinon_1 = __importDefault(require("sinon"));
const src_1 = require("../../src");
const utils_1 = __importDefault(require("../../src/lib/utils"));
const injective_1 = require("../resources/injective");
const should = require("should");
const nock = require("nock");
describe('INJ', function () {
    let bitgo;
    let basecoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('injective', src_1.Injective.createInstance);
        bitgo.safeRegister('tinjective', src_1.Tinjective.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('tinjective');
    });
    it('should return the right info', function () {
        const injective = bitgo.coin('injective');
        const tinjective = bitgo.coin('tinjective');
        injective.getChain().should.equal('injective');
        injective.getFamily().should.equal('injective');
        injective.getFullName().should.equal('Injective');
        injective.getBaseFactor().should.equal(1e18);
        tinjective.getChain().should.equal('tinjective');
        tinjective.getFamily().should.equal('injective');
        tinjective.getFullName().should.equal('Testnet Injective');
        tinjective.getBaseFactor().should.equal(1e18);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const addressDetails = basecoin.getAddressDetails(injective_1.address.noMemoIdAddress);
            addressDetails.address.should.equal(injective_1.address.noMemoIdAddress);
            should.not.exist(addressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const addressDetails = basecoin.getAddressDetails(injective_1.address.validMemoIdAddress);
            addressDetails.address.should.equal(injective_1.address.validMemoIdAddress.split('?')[0]);
            addressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                basecoin.getAddressDetails(injective_1.address.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const receiveAddress = {
                address: 'inj1vftzm2nnlt0x8z9gq2hgxxywlk79m8k7k3ljgk?memoId=7',
                coinSpecific: {
                    rootAddress: 'inj1vftzm2nnlt0x8z9gq2hgxxywlk79m8k7k3ljgk',
                    memoID: '7',
                },
            };
            const isValid = await basecoin.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(utils_1.default.isValidAddress(injective_1.address.address1), true);
            should.equal(utils_1.default.isValidAddress(injective_1.address.address2), true);
            should.equal(utils_1.default.isValidAddress(injective_1.address.address3), false);
            should.equal(utils_1.default.isValidAddress(injective_1.address.address4), false);
            should.equal(utils_1.default.isValidAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidAddress(undefined), false);
            should.equal(utils_1.default.isValidAddress(''), false);
            should.equal(utils_1.default.isValidAddress(injective_1.address.validMemoIdAddress), true);
            should.equal(utils_1.default.isValidAddress(injective_1.address.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(utils_1.default.isValidValidatorAddress(injective_1.address.validatorAddress1), true);
            should.equal(utils_1.default.isValidValidatorAddress(injective_1.address.validatorAddress2), true);
            should.equal(utils_1.default.isValidValidatorAddress(injective_1.address.validatorAddress3), false);
            should.equal(utils_1.default.isValidValidatorAddress(injective_1.address.validatorAddress4), false);
            should.equal(utils_1.default.isValidValidatorAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidValidatorAddress(undefined), false);
            should.equal(utils_1.default.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: injective_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: injective_1.TEST_SEND_TX.recipient,
                        amount: injective_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify delegate transaction', async function () {
            const txPrebuild = {
                txHex: injective_1.TEST_DELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: injective_1.TEST_DELEGATE_TX.validator,
                        amount: injective_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify undelegate transaction', async function () {
            const txPrebuild = {
                txHex: injective_1.TEST_UNDELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: injective_1.TEST_UNDELEGATE_TX.validator,
                        amount: injective_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify withdraw rewards transaction', async function () {
            const txPrebuild = {
                txHex: injective_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: injective_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: injective_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: injective_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: injective_1.TEST_SEND_TX.recipient,
                        amount: injective_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: injective_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: injective_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a delegate transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: injective_1.TEST_DELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: injective_1.TEST_DELEGATE_TX.hash,
                outputs: [
                    {
                        address: injective_1.TEST_DELEGATE_TX.validator,
                        amount: injective_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: injective_1.TEST_DELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: injective_1.TEST_DELEGATE_TX.gasBudget.amount[0].amount },
                type: 13,
            });
        });
        it('should explain a undelegate transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: injective_1.TEST_UNDELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: injective_1.TEST_UNDELEGATE_TX.hash,
                outputs: [
                    {
                        address: injective_1.TEST_UNDELEGATE_TX.validator,
                        amount: injective_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: injective_1.TEST_UNDELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: injective_1.TEST_UNDELEGATE_TX.gasBudget.amount[0].amount },
                type: 17,
            });
        });
        it('should explain a withdraw transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: injective_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: injective_1.TEST_WITHDRAW_REWARDS_TX.hash,
                outputs: [
                    {
                        address: injective_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
                outputAmount: undefined,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: injective_1.TEST_WITHDRAW_REWARDS_TX.gasBudget.amount[0].amount },
                type: 15,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: injective_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: injective_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: injective_1.TEST_TX_WITH_MEMO.to,
                        amount: injective_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: injective_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: injective_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: injective_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await basecoin.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await basecoin.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        it('should parse a transfer transaction', async function () {
            const transferInputsResponse = {
                address: injective_1.TEST_SEND_TX.recipient,
                amount: new bignumber_js_1.default(injective_1.TEST_SEND_TX.sendAmount).plus(injective_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
            };
            const transferOutputsResponse = {
                address: injective_1.TEST_SEND_TX.recipient,
                amount: injective_1.TEST_SEND_TX.sendAmount,
            };
            const parsedTransaction = await basecoin.parseTransaction({ txHex: injective_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Injective.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await basecoin
                .parseTransaction({ txHex: injective_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
    describe('Recover transaction: success path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = injective_1.wrwUser.destinationAddress;
        const coin = statics_1.coins.get('tinjective');
        const testBalance = [
            {
                denom: 'inj',
                amount: '15000000000000000',
            },
        ];
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            nock('https://testnet.sentry.lcd.injective.network')
                .get('/cosmos/auth/v1beta1/accounts/' + injective_1.wrwUser.senderAddress)
                .reply(200, injective_1.mockAccountDetailsResponse);
            const accountBalance = sandBox.stub(src_1.Injective.prototype, 'getAccountBalance');
            accountBalance.withArgs(injective_1.wrwUser.senderAddress).resolves(testBalance);
            const chainId = sandBox.stub(src_1.Injective.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should recover funds for non-bitgo recoveries', async function () {
            const res = await basecoin.recover({
                userKey: injective_1.wrwUser.userPrivateKey,
                backupKey: injective_1.wrwUser.backupPrivateKey,
                bitgoKey: injective_1.wrwUser.bitgoPublicKey,
                walletPassphrase: injective_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(basecoin.getAccountBalance);
            sandBox.assert.calledOnce(basecoin.getChainId);
            const txn = new abstract_cosmos_1.CosmosTransaction(coin, utils_1.default);
            txn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const txnJson = txn.toJson();
            const sendMessage = txnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const gasAmount = new bignumber_js_1.default(125000000000000);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.toAddress, destinationAddress);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
            should.equal(sendMessage.amount[0].denom, testBalance[0].denom);
        });
    });
    describe('Recover transaction: failure path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = injective_1.wrwUser.destinationAddress;
        const testZeroBalance = [
            {
                denom: 'inj',
                amount: '0',
            },
        ];
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            nock('https://testnet.sentry.lcd.injective.network')
                .get('/cosmos/auth/v1beta1/accounts/' + injective_1.wrwUser.senderAddress)
                .reply(200, injective_1.mockAccountDetailsResponse);
            const accountBalance = sandBox.stub(src_1.Injective.prototype, 'getAccountBalance');
            accountBalance.withArgs(injective_1.wrwUser.senderAddress).resolves(testZeroBalance);
            const chainId = sandBox.stub(src_1.Injective.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should throw error if backupkey is not present', async function () {
            await basecoin
                .recover({
                userKey: injective_1.wrwUser.userPrivateKey,
                bitgoKey: injective_1.wrwUser.bitgoPublicKey,
                walletPassphrase: injective_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing backupKey');
        });
        it('should throw error if userkey is not present', async function () {
            await basecoin
                .recover({
                backupKey: injective_1.wrwUser.backupPrivateKey,
                bitgoKey: injective_1.wrwUser.bitgoPublicKey,
                walletPassphrase: injective_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing userKey');
        });
        it('should throw error if wallet passphrase is not present', async function () {
            await basecoin
                .recover({
                userKey: injective_1.wrwUser.userPrivateKey,
                backupKey: injective_1.wrwUser.backupPrivateKey,
                bitgoKey: injective_1.wrwUser.bitgoPublicKey,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing wallet passphrase');
        });
        it('should throw error if there is no balance', async function () {
            await basecoin
                .recover({
                userKey: injective_1.wrwUser.userPrivateKey,
                backupKey: injective_1.wrwUser.backupPrivateKey,
                bitgoKey: injective_1.wrwUser.bitgoPublicKey,
                walletPassphrase: injective_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('Did not have enough funds to recover');
        });
    });
});
