// @bitgo/abstract-utxo/dist/esm/descriptor/NamedDescriptor.js
import * as t from 'io-ts';
import { Descriptor, message } from '@bitgo/wasm-utxo';
export const NamedDescriptor = t.intersection([
    t.type({
        name: t.string,
        value: t.string,
    }),
    t.partial({
        signatures: t.union([t.array(t.string), t.undefined]),
    }),
], 'NamedDescriptor');
export function createNamedDescriptorWithSignature(name, descriptor, signingKey) {
    if (typeof descriptor === 'string') {
        descriptor = Descriptor.fromString(descriptor, 'derivable');
    }
    const value = descriptor.toString();
    const signature = Buffer.from(message.signMessage(value, signingKey.privateKey)).toString('hex');
    return { name, value, signatures: [signature] };
}
export function toNamedDescriptorNative(e, pkType) {
    return { ...e, value: Descriptor.fromString(e.value, pkType) };
}
export function hasValidSignature(descriptor, key, signatures) {
    if (typeof descriptor === 'string') {
        descriptor = Descriptor.fromString(descriptor, 'derivable');
    }
    const descriptorString = descriptor.toString();
    return signatures.some((signature) => {
        return message.verifyMessage(descriptorString, key.publicKey, Buffer.from(signature, 'hex'));
    });
}
export function assertHasValidSignature(namedDescriptor, key) {
    if (!hasValidSignature(namedDescriptor.value, key, namedDescriptor.signatures ?? [])) {
        throw new Error(`Descriptor ${namedDescriptor.name} does not have a valid signature (key=${key.toBase58()})`);
    }
}
