// @bitgo/abstract-utxo/dist/esm/descriptor/validatePolicy.js
import { descriptorWallet } from '@bitgo/wasm-utxo';
import { parseDescriptor } from './builder';
import { hasValidSignature, toNamedDescriptorNative } from './NamedDescriptor';
export const policyAllowAll = {
    name: 'allowAll',
    validate: () => true,
};
export function getValidatorDescriptorTemplate(name) {
    return {
        name: 'descriptorTemplate(' + name + ')',
        validate(arr, walletKeys) {
            return arr.every((d) => {
                const parsed = parseDescriptor(d.value);
                return (parsed.name === name &&
                    parsed.keys.length === walletKeys.length &&
                    parsed.keys.every((k, i) => k.toBase58() === walletKeys[i].neutered().toBase58()));
            });
        },
    };
}
export function getValidatorEvery(validators) {
    return {
        name: 'every(' + validators.map((v) => v.name).join(',') + ')',
        validate(arr, walletKeys) {
            return validators.every((v) => v.validate(arr, walletKeys));
        },
    };
}
export function getValidatorSome(validators) {
    return {
        name: 'some(' + validators.map((v) => v.name).join(',') + ')',
        validate(arr, walletKeys) {
            return validators.some((v) => v.validate(arr, walletKeys));
        },
    };
}
export function getValidatorOneOfTemplates(names) {
    return getValidatorSome(names.map(getValidatorDescriptorTemplate));
}
export function getValidatorSignedByUserKey() {
    return {
        name: 'signedByUser',
        validate(arr, walletKeys) {
            // the first key is the user key, by convention
            return arr.every((d) => hasValidSignature(d.value, walletKeys[0], d.signatures ?? []));
        },
    };
}
export class DescriptorPolicyValidationError extends Error {
    constructor(ds, policy) {
        super(`Descriptors ${ds.map((d) => d.value.toString())} does not match policy ${policy.name}`);
    }
}
export function assertDescriptorPolicy(descriptors, policy, walletKeys) {
    if (!policy.validate(descriptors, walletKeys)) {
        throw new DescriptorPolicyValidationError(descriptors, policy);
    }
}
export function toDescriptorMapValidate(descriptors, walletKeys, policy) {
    const namedDescriptorsNative = descriptors.map((v) => toNamedDescriptorNative(v, 'derivable'));
    assertDescriptorPolicy(namedDescriptorsNative, policy, walletKeys);
    return descriptorWallet.toDescriptorMap(namedDescriptorsNative);
}
export function getPolicyForEnv(env) {
    switch (env) {
        case 'adminProd':
        case 'prod':
            return getValidatorSome([
                // allow 2-of-3-ish descriptor groups where the keys match the wallet keys
                getValidatorDescriptorTemplate('Wsh2Of3'),
                // allow descriptor groups where all keys match the wallet keys plus OP_DROP (coredao staking)
                getValidatorDescriptorTemplate('Wsh2Of3CltvDrop'),
                // allow all descriptors signed by the user key
                getValidatorSignedByUserKey(),
            ]);
        default:
            return policyAllowAll;
    }
}
