// @bitgo/abstract-utxo/dist/cjs/test/unit/transaction/descriptor/parse.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const testutils = __importStar(require("@bitgo/wasm-utxo/testutils"));
const parse_1 = require("../../../../src/transaction/descriptor/parse");
const verifyTransaction_1 = require("../../../../src/transaction/descriptor/verifyTransaction");
const parseToAmountType_1 = require("../../../../src/transaction/descriptor/parseToAmountType");
const fixtures_utils_1 = require("./fixtures.utils");
const { getDefaultXPubs, getDescriptor, getDescriptorMap, mockPsbtDefault } = testutils.descriptor;
const { toPlainObject } = testutils;
const { assertEqualFixture } = (0, fixtures_utils_1.getFixtureRoot)(__dirname + '/fixtures');
function toBaseOutput(output, amountType) {
    (0, assert_1.default)(output.address);
    return {
        address: output.address,
        amount: (0, parseToAmountType_1.toAmountType)(output.value, amountType),
    };
}
function toBaseOutputBigInt(output) {
    return toBaseOutput(output, 'bigint');
}
function toBaseOutputString(output) {
    return toBaseOutput(output, 'string');
}
function toMaxOutput(output) {
    return {
        ...output,
        value: 'max',
    };
}
describe('parse', function () {
    const descriptorSelf = getDescriptor('Wsh2Of3', getDefaultXPubs('a'));
    const descriptorOther = getDescriptor('Wsh2Of3', getDefaultXPubs('b'));
    const psbt = mockPsbtDefault({ descriptorSelf, descriptorOther });
    function recipient(descriptor, index, value = 1000) {
        return { value, address: wasm_utxo_1.descriptorWallet.createAddressFromDescriptor(descriptor, index, 'btc') };
    }
    function internalRecipient(index, value) {
        return recipient(descriptorSelf, index, value);
    }
    function externalRecipient(index, value) {
        return recipient(descriptorOther, index, value);
    }
    function getBaseParsedTransaction(psbt, recipients) {
        return (0, parse_1.toBaseParsedTransactionOutputsFromPsbt)(psbt, getDescriptorMap('Wsh2Of3', getDefaultXPubs('a')), recipients.map(toBaseOutputString), 'btc');
    }
    const psbtOutputs = psbt.getOutputsWithAddress('btc');
    const psbtOutput0 = psbtOutputs[0];
    const psbtOutput1 = psbtOutputs[1];
    describe('toBase', function () {
        it('should return the correct BaseParsedTransactionOutputs', async function () {
            await assertEqualFixture('parseWithoutRecipients.json', toPlainObject(getBaseParsedTransaction(psbt, [])));
            await assertEqualFixture('parseWithExternalRecipient.json', toPlainObject(getBaseParsedTransaction(psbt, [psbtOutput0])));
            await assertEqualFixture('parseWithInternalRecipient.json', toPlainObject(getBaseParsedTransaction(psbt, [psbtOutput1])));
            await assertEqualFixture('parseWithExternalRecipient.json', 
            // max recipient: ignore actual value
            toPlainObject(getBaseParsedTransaction(psbt, [toMaxOutput(psbtOutput0)])));
        });
        function assertEqualValidationError(actual, expected) {
            function normErrors(e) {
                return e.map((e) => ({ ...e, stack: undefined }));
            }
            if (actual instanceof verifyTransaction_1.AggregateValidationError) {
                assert_1.default.deepStrictEqual(normErrors(actual.errors), normErrors(expected.errors));
            }
            else {
                throw new Error('unexpected error type: ' + actual);
            }
        }
        function assertValidationError(f, expected) {
            assert_1.default.throws(f, (err) => {
                assertEqualValidationError(err, expected);
                return true;
            });
        }
        function implicitOutputError(output, { external = true } = {}) {
            return new verifyTransaction_1.ErrorImplicitExternalOutputs([{ ...toBaseOutputBigInt(output), external }]);
        }
        function missingOutputError(output, { external = true } = {}) {
            return new verifyTransaction_1.ErrorMissingOutputs([{ ...toBaseOutputBigInt(output), external }]);
        }
        it('should throw expected error: no recipient requested', function () {
            assertValidationError(() => (0, verifyTransaction_1.assertExpectedOutputDifference)(getBaseParsedTransaction(psbt, [])), new verifyTransaction_1.AggregateValidationError([implicitOutputError(psbtOutput0)]));
        });
        it('should throw expected error: only internal recipient requested', function () {
            assertValidationError(() => (0, verifyTransaction_1.assertExpectedOutputDifference)(getBaseParsedTransaction(psbt, [psbtOutput1])), new verifyTransaction_1.AggregateValidationError([implicitOutputError(psbtOutput0)]));
        });
        it('should throw expected error: only internal max recipient requested', function () {
            assertValidationError(() => (0, verifyTransaction_1.assertExpectedOutputDifference)(getBaseParsedTransaction(psbt, [toMaxOutput(psbtOutput1)])), new verifyTransaction_1.AggregateValidationError([implicitOutputError(psbtOutput0)]));
        });
        it('should throw expected error: swapped recipient', function () {
            const recipient = externalRecipient(99);
            assertValidationError(() => (0, verifyTransaction_1.assertExpectedOutputDifference)(getBaseParsedTransaction(psbt, [recipient])), new verifyTransaction_1.AggregateValidationError([missingOutputError(recipient), implicitOutputError(psbtOutput0)]));
        });
        it('should throw expected error: missing internal recipient', function () {
            const recipient = internalRecipient(99);
            assertValidationError(() => (0, verifyTransaction_1.assertExpectedOutputDifference)(getBaseParsedTransaction(psbt, [recipient])), new verifyTransaction_1.AggregateValidationError([missingOutputError(recipient), implicitOutputError(psbtOutput0)]));
        });
    });
});
