// @bitgo/sdk-coin-stx/dist/test/unit/sip10Token.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = __importDefault(require("lodash"));
const nock_1 = __importDefault(require("nock"));
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../src");
const testData = __importStar(require("../fixtures"));
const assert_1 = __importDefault(require("assert"));
describe('Sip10Token:', function () {
    const sip10TokenName = 'tstx:tsip6dp';
    let bitgo;
    let basecoin;
    let newTxPrebuild;
    let newTxParams;
    let wallet;
    const txPreBuild = {
        txHex: testData.txForExplainFungibleTokenTransfer,
        txInfo: {},
    };
    const txParams = {
        recipients: testData.recipients,
    };
    const memo = {
        type: '',
        value: '1',
    };
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, {
            env: 'mock',
        });
        bitgo.initializeTestVars();
        src_1.Sip10Token.createTokenConstructors().forEach(({ name, coinConstructor }) => {
            bitgo.safeRegister(name, coinConstructor);
        });
        newTxPrebuild = () => {
            return lodash_1.default.cloneDeep(txPreBuild);
        };
        newTxParams = () => {
            return lodash_1.default.cloneDeep(txParams);
        };
        basecoin = bitgo.coin(sip10TokenName);
        wallet = new sdk_core_1.Wallet(bitgo, basecoin, {});
    });
    describe('Verify Transaction', function () {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.memo = memo;
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams,
                txPrebuild,
                verification,
                wallet,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify transaction when recipients amount are numbers', async function () {
            const txPrebuild = newTxPrebuild();
            const txParamsWithNumberAmounts = newTxParams();
            txParamsWithNumberAmounts.recipients = txParamsWithNumberAmounts.recipients.map(({ address, amount, memo, tokenName }) => {
                return { address, amount: Number(amount), memo, tokenName };
            });
            txParamsWithNumberAmounts.memo = memo;
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams: txParamsWithNumberAmounts,
                txPrebuild,
                verification,
                wallet,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify when memo is passed', async function () {
            const txPrebuild = newTxPrebuild();
            txPrebuild.txHex = testData.txForExplainFungibleTokenTransferWithMemoId10;
            const txParams = newTxParams();
            const verification = {};
            txParams.memo = {
                type: '',
                value: '10',
            };
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams: txParams,
                txPrebuild,
                verification,
                wallet,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify when memo is zero', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txPrebuild.txHex = testData.txForExplainFungibleTokenTransferWithMemoZero;
            txParams.memo = {
                type: '',
                value: '0',
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams: txParams,
                txPrebuild,
                verification,
                wallet,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify when memo is passed inside recipient address', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.recipients[0].address = 'SN2NN1JP9AEP5BVE19RNJ6T2MP7NDGRZYST1VDF3M?memoId=10';
            txPrebuild.txHex = testData.txForExplainFungibleTokenTransferWithMemoId10;
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams: txParams,
                txPrebuild,
                verification,
                wallet,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify when memo is not passed', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txPrebuild.txHex = testData.txForExplainFungibleTokenTransferWithoutMemo;
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({
                txParams: txParams,
                txPrebuild,
                verification,
                wallet,
            });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with no recipients', async function () {
            const txPrebuild = {};
            const txParams = newTxParams();
            txParams.recipients = [];
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
                wallet,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
        it('should fail when more than 1 recipients are passed', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.recipients.push({
                address: 'SN2NN1JP9AEP5BVE19RNJ6T2MP7NDGRZYST1VDF3N',
                amount: '10000',
                memo: '1',
                tokenName: 'tsip6dp-token',
            });
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
                wallet,
            })
                .should.rejectedWith("tstx:tsip6dp doesn't support sending to more than 1 destination address within a single transaction. Try again, using only a single recipient.");
        });
        it('should fail to verify transaction with wrong address', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.recipients[0].address = 'SN2NN1JP9AEP5BVE19RNJ6T2MP7NDGRZYST1VDF3N';
            const verification = {};
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
                verification,
                wallet,
            })
                .should.rejectedWith('Tx outputs does not match with expected txParams recipients');
        });
        it('should fail to verify transaction with wrong amount', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.recipients[0].amount = '100';
            const verification = {};
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
                verification,
                wallet,
            })
                .should.rejectedWith('Tx outputs does not match with expected txParams recipients');
        });
        it('should fail to verify transaction with wrong memo', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.recipients[0].memo = '2';
            const verification = {};
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
                verification,
                wallet,
            })
                .should.rejectedWith('Tx memo does not match with expected txParams recipient memo');
        });
        it('should fail to verify transaction with wrong token', async function () {
            const txPrebuild = newTxPrebuild();
            const txParams = newTxParams();
            txParams.recipients[0].tokenName = 'tstx:tsip8dp';
            const verification = {};
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
                verification,
                wallet,
            })
                .should.rejectedWith('Tx outputs does not match with expected txParams recipients');
        });
    });
    describe('Recover Transaction SIP10', () => {
        before(function () {
            nock_1.default.enableNetConnect();
        });
        beforeEach(function () {
            nock_1.default.cleanAll();
        });
        after(function () {
            nock_1.default.disableNetConnect();
        });
        it('should build a signed token recover transaction when private key data is passed', async () => {
            const rootAddress = testData.HOT_WALLET_ROOT_ADDRESS;
            (0, nock_1.default)(`https://api.testnet.hiro.so`)
                .get(`/extended/v2/addresses/${rootAddress}/balances/stx`)
                .reply(200, testData.ACCOUNT_BALANCE_RESPONSE);
            (0, nock_1.default)('https://api.testnet.hiro.so')
                .get(`/extended/v2/addresses/${rootAddress}/balances/ft/${testData.STX_TOKEN_ASSET_ID}`)
                .reply(200, testData.TOKEN_BALANCE_RESPONSE);
            (0, nock_1.default)(`https://api.testnet.hiro.so`)
                .get(`/extended/v1/address/${rootAddress}/nonces`)
                .reply(200, testData.ACCOUNT_NONCE_RESPONSE);
            (0, nock_1.default)(`https://api.testnet.hiro.so`, { allowUnmocked: true })
                .post(`/v2/fees/transaction`, testData.FEE_ESTIMATION_TOKEN_REQUEST)
                .reply(200, testData.FEE_ESTIMATION_TOKEN_RESPONSE);
            const recoveryOptions = {
                backupKey: testData.HOT_WALLET_KEY_CARD_INFO.BACKUP_KEY,
                userKey: testData.HOT_WALLET_KEY_CARD_INFO.USER_KEY,
                rootAddress: rootAddress,
                recoveryDestination: testData.DESTINATION_ADDRESS_WRW,
                bitgoKey: testData.HOT_WALLET_KEY_CARD_INFO.BITGO_PUB_KEY,
                walletPassphrase: testData.HOT_WALLET_KEY_CARD_INFO.WALLET_PASSPHRASE,
                contractId: 'STAG18E45W613FZ3H4ZMF6QHH426EXM5QTSAVWYH.tsip6dp-token',
            };
            const response = await basecoin.recover(recoveryOptions);
            response.should.have.property('txHex');
            assert_1.default.deepEqual(response.txHex, testData.HOT_WALLET_TOKEN_RECOVERY_TX_HEX, 'tx hex not matching!');
        });
        it('should build an unsigned token transaction when public keys are passed', async () => {
            const rootAddress = testData.COLD_WALLET_ROOT_ADDRESS;
            (0, nock_1.default)(`https://api.testnet.hiro.so`)
                .get(`/extended/v2/addresses/${rootAddress}/balances/stx`)
                .reply(200, testData.ACCOUNT_BALANCE_RESPONSE);
            (0, nock_1.default)('https://api.testnet.hiro.so')
                .get(`/extended/v2/addresses/${rootAddress}/balances/ft/${testData.STX_TOKEN_ASSET_ID}`)
                .reply(200, testData.TOKEN_BALANCE_RESPONSE);
            (0, nock_1.default)(`https://api.testnet.hiro.so`, { allowUnmocked: true })
                .get(`/extended/v1/address/${rootAddress}/nonces`)
                .reply(200, testData.ACCOUNT_NONCE_RESPONSE);
            const feeEstimateRequest = testData.FEE_ESTIMATION_TOKEN_REQUEST;
            feeEstimateRequest.transaction_payload =
                '021a1500a1c42f0c11bfe3893f479af18904677685be0d747369703664702d746f6b656e087472616e73666572000000040100000000000000000000000005f5e10005159f2f1aff6fa0062e1f7fa6096133e75f47a7e8f7051a1500a1c42f0c11bfe3893f479af18904677685be09';
            (0, nock_1.default)(`https://api.testnet.hiro.so`)
                .post(`/v2/fees/transaction`, feeEstimateRequest)
                .reply(200, testData.FEE_ESTIMATION_TOKEN_RESPONSE);
            const recoveryOptions = {
                backupKey: testData.COLD_WALLET_PUBLIC_KEY_INFO.BACKUP_KEY,
                userKey: testData.COLD_WALLET_PUBLIC_KEY_INFO.USER_KEY,
                rootAddress: rootAddress,
                recoveryDestination: testData.DESTINATION_ADDRESS_WRW,
                bitgoKey: testData.COLD_WALLET_PUBLIC_KEY_INFO.BITGO_PUB_KEY,
                contractId: 'STAG18E45W613FZ3H4ZMF6QHH426EXM5QTSAVWYH.tsip6dp-token',
            };
            const response = (await basecoin.recover(recoveryOptions));
            response.should.have.property('txHex');
            response.should.have.property('coin');
            response.should.have.property('feeInfo');
            assert_1.default.deepEqual(response.txHex, testData.COLD_WALLET_TOKEN_UNSIGNED_SWEEP_TX_HEX, 'tx hex not matching!');
            assert_1.default.deepEqual(response.coin, 'tstx:tsip6dp', 'coin not matching!');
        });
        it('should fail with insufficient balance when native stx balance is lower than fee for sip10', async () => {
            const rootAddress = testData.HOT_WALLET_ROOT_ADDRESS;
            const accountBalance = JSON.parse(JSON.stringify(testData.ACCOUNT_BALANCE_RESPONSE));
            accountBalance.balance = '100'; // set balance lower than fee
            (0, nock_1.default)(`https://api.testnet.hiro.so`)
                .get(`/extended/v2/addresses/${rootAddress}/balances/stx`)
                .reply(200, accountBalance);
            (0, nock_1.default)('https://api.testnet.hiro.so')
                .get(`/extended/v2/addresses/${rootAddress}/balances/ft/${testData.STX_TOKEN_ASSET_ID}`)
                .reply(200, testData.TOKEN_BALANCE_RESPONSE);
            (0, nock_1.default)(`https://api.testnet.hiro.so`)
                .get(`/extended/v1/address/${rootAddress}/nonces`)
                .reply(200, testData.ACCOUNT_NONCE_RESPONSE);
            const feeRequestBody = testData.FEE_ESTIMATION_TOKEN_REQUEST;
            feeRequestBody.transaction_payload =
                '021a1500a1c42f0c11bfe3893f479af18904677685be0d747369703664702d746f6b656e087472616e73666572000000040100000000000000000000000005f5e100051549857eb4b6dd4fee08c3ec04e3d0ed04ef67d324051a1500a1c42f0c11bfe3893f479af18904677685be09';
            (0, nock_1.default)(`https://api.testnet.hiro.so`, { allowUnmocked: true })
                .post(`/v2/fees/transaction`, feeRequestBody)
                .reply(200, testData.FEE_ESTIMATION_TOKEN_RESPONSE);
            const recoveryOptions = {
                backupKey: testData.HOT_WALLET_KEY_CARD_INFO.BACKUP_KEY,
                userKey: testData.HOT_WALLET_KEY_CARD_INFO.USER_KEY,
                rootAddress: rootAddress,
                recoveryDestination: testData.DESTINATION_ADDRESS_WRW,
                bitgoKey: testData.HOT_WALLET_KEY_CARD_INFO.BITGO_PUB_KEY,
                walletPassphrase: testData.HOT_WALLET_KEY_CARD_INFO.WALLET_PASSPHRASE,
                contractId: 'STAG18E45W613FZ3H4ZMF6QHH426EXM5QTSAVWYH.tsip6dp-token',
            };
            await basecoin.recover(recoveryOptions).should.rejectedWith('insufficient balance to build the transaction');
        });
        it('should fail when only contract address is passed', async () => {
            const recoveryOptions = {
                backupKey: testData.HOT_WALLET_KEY_CARD_INFO.BACKUP_KEY,
                userKey: testData.HOT_WALLET_KEY_CARD_INFO.USER_KEY,
                rootAddress: testData.HOT_WALLET_ROOT_ADDRESS,
                recoveryDestination: testData.DESTINATION_ADDRESS_WRW,
                bitgoKey: testData.HOT_WALLET_KEY_CARD_INFO.BITGO_PUB_KEY,
                walletPassphrase: testData.HOT_WALLET_KEY_CARD_INFO.WALLET_PASSPHRASE,
                contractId: 'STAG18E45W613FZ3H4ZMF6QHH426EXM5QTSAVWYH',
            };
            await basecoin
                .recover(recoveryOptions)
                .should.rejectedWith('invalid contract id, please provide it in the form (contractAddress.contractName)');
        });
    });
});
