// @bitgo/sdk-coin-tao/dist/test/unit/fixtures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.westendBlock = exports.consolidationWrwUser = exports.unsignedSweepUser = exports.wrwUser = void 0;
exports.wrwUser = {
    userKey: '{"iv":"MceJIaIaL2D37g7xo9XrlQ==","v":1,"iter":10000,"ks":256,"ts":64,"mode"\n' +
        ':"ccm","adata":"","cipher":"aes","salt":"30XLHXfzATg=","ct":"0cx4eIXluoqdA8\n' +
        'cxzCcvOie8oBThg0wRU/G+ix3V/Wj/mzVy3Nc4MQOLnISE7F7CkepwTZnAYezjsIjFktrdxwg6g\n' +
        'M54w+nI9k2C1uc5MV2XvmJ6NU84R4g8tQ58QQtg5dwbxvppQ4sZbCivU+fsVoB/mrg09+oaTGA5\n' +
        'wtLyHbuSFBAYsmysJzuGsWIXPu0yxFBo5FigcBqqb2alI4AwCrztPVfK8jbk+p44BOHH6rn5KYw\n' +
        'L8jGyPffxZOC16EP7yuhze+D4jmyRuBFqQYKra10hYge5mfxW/0DP1C/lLuXDYljcgxX5Oesezw\n' +
        'UegwXxel7WgjUWfbWG12wnqo6YEd0LuSXVC7l/3I4Oxhz06cgL+Kr5Wj3XyjSPvkMjquIRe8xHl\n' +
        'LQqy9ubtkhB66k+Yonx4jjcELBzie9FM+oFfTqWseKAh1EDapR5stwBEVQ5ZaQmD39Q531YG7pI\n' +
        'zgShnnd8Kk9zmVR6OzHKhIW0BmmAhui+DDE2Zo9e1WiqoVb7V+qs2BAm5MKpUOpU5QCjmKe7wDR\n' +
        'H4f2j/sTSPLfxdeRJyscpBjoFbdk+nyIQFCB7nWApCd269I2rH4pYN6o1BubKkdI8x0SNoFpSvu\n' +
        'dv2d+0DRXZd8w01h1+HryHtn5ZOzOKx0lWJZyDir/lRayL+EqZAN+HsceOmZmbRV9l+A4BFWJXv\n' +
        'pgDpCEaFgMPX1CQXbpdFVDSZDq61Mepf+qh0rCyK78bGDsuFr2qomE9VNYftY5oJdLSdiTUT6CO\n' +
        'DYN6birj9RbDCPQRGSqaALTc4kHnl4BiT14TZyBd4mH41jWvvAK0j2h8hrhRiWCKmSoFSJ5gR5r\n' +
        '56aHWXxmW5zTZNypF6aP+yavyJ7iTeajETJrdFLE2Ey998/1TVTgYd79vZqMe+IYrzvBNXb2wRY\n' +
        'LwEPDdNr8a8vGim/H7baosOZjDqAv4FUUC+aZKikCxnhm8A9m+QtlU0BeenUaQQ27V0MF8aFk4M\n' +
        'a6fw0k963BZDtLw1ZXJ40rsn5OGpuzJwdKvW71XnnQ3RVec7NFTZ7gHzYj05R9Qk8od0oHGNRE1\n' +
        '0hU7OmXjo0K21sTlGF1YvfjH/sw8DoY9O/soNjOohk9w5YrM191vDsZEOeY9RbcKO8QW1BapjoB\n' +
        'jW3bAEKBuQiU5MD/IanHKI8wpcjKBtPY+VIR01sms0UJvdLlulyHgfavYK4I81yzs"}',
    backupKey: '{"iv":"QFWrc1AcWXiiNjcS/d6XwA==","v":1,"iter":10000,"ks":256,"ts":64,"mode"\n' +
        ':"ccm","adata":"","cipher":"aes","salt":"yMEERi7I+zI=","ct":"gjuxyv0KZka509\n' +
        'xc8ZZ5WID8l0i7yVcq7OOWrKTplYQurJCu3Sx5KQtwxQaTEJ1JK5CnJqk917rDkuzbHruez0WOF\n' +
        'jHTIEkWOwOanvfB68hVjkB5naoxq8bM1/mk19Gy5drY3txtiP/kPdtgwsOWwVKWHEAiz38ksPw8\n' +
        '+WpwHfTaCEC4HLl/cwQx9bDoL7VeBAUqCkBayanGTwgI37HOp+DySrzYWOkvFJqsa8s7PVWN3fR\n' +
        'OpB7Ia9D/iJCjaPcvRA6U9qd730hCH3IfV7pqHGJkAurPAUqOEJaDVXn4mzfRkE7iKrPGAyJdYt\n' +
        'jrAw3JZ1hYC1WiJ30+290GPH4mStCz+sfHnZI/N+Ot9clUViBN8FvZDrG5Yvspzn9cm2Azv2l+x\n' +
        'kwAzOO2Raq4buwEumFXA+2sy8Q/zy3+zdij1KECfyF24N9T+gwSEBdu3LrjoVqQpzKP8v+oR5qL\n' +
        '4HQ3jd1evLFd6k5uFIeYq9MmpLh0JPfabQlG2ZeFK5Xf2AESCvCCjhvNAISUkYYaQ8Ej6QV/HLX\n' +
        '1AHpyPpW4u3dbZ8xmhskw6vp15VWhBvizyAU1Oj29x4lYtYjAHjbELrK9WxndTWvDyYv8L7MjlQ\n' +
        'ThOnWkaKUgIGsow2PdiOEa4LC8qszpRoaaqqewbVywy3CWMhuaCrv/wLs+aaG8TSmJhpNYZOVIN\n' +
        'hcT2BGIGXh4+u1nzWZrVkQ6D/06YESvQLn/XmwFklWCOXKSVAFW+qIehhL4pVtF3pzT7SY1FQGu\n' +
        'GeW1NTEtiiL9e4VfOr7beNY/rL7lxcEqxz4BNUcGf3q2fj/7xRNAy3jw74amgszKuTlkkVCrgyl\n' +
        '1jScd+5ZC4P58mNuR0sTlPq4sWKVRYiVIcptbGbQ0vPaA7D/L0Ef1S4KHdDgwFRSEggjJylzLwW\n' +
        '/Z/LdreZAvPIvW/M/z0BRbD8e3AHl1/oHvSRpptW221uIBbULEVq3gPU/xLY+eAGmtuFv+Qs4bb\n' +
        's9kUqPmwTZnrPsHSw1L2mGZ2aj5YTeZwyFep1KOgr3exVz4trSkCPQI0ON97GjY3d+bKsB2ae4q\n' +
        'DguXb3oAx2bLbklRk94zMsW92Xl3wMM00ywpxBij7wSeL79qRRATcF08oU6EaEaVeF/g+J1Ptnv\n' +
        'iGfvk6RFq14c+oTgWZI8GT5pljqgo8VcCo6tyGGfKLph6yOhUBvz8UpKVCJRnFA=="}',
    bitgoKey: '477997ac5b6ade94fffecfbaf7663d6880f19dd40a6b3da5a45e585132efb38a67707eb050d\n' +
        '5219984aceaa39115da17439bb9cd0a8a0387ec1dc47843efd882',
    walletPassphrase: '#Bondiola1234',
    walletAddress0: '5HCqqv9fiQpq7XFuyWfEMGACahQnZrYig6Ac7JigGikX8YWs',
    walletAddress1: '5CBS7Mtz53gu3zQGxw3a21CazJDd2ESGKgZWtBUran3iGQ2X',
};
exports.unsignedSweepUser = {
    userKey: 'd4a570c1831c7024cb72533e42690cec9d9150b9aac59001c4a5a3a285518263a9e89f1c709\n' +
        'bcb1e832f2ff5169fecd496881ba8970f69d5adc977ba4799fedb',
    backupKey: 'd4a570c1831c7024cb72533e42690cec9d9150b9aac59001c4a5a3a285518263a9e89f1c709\n' +
        'bcb1e832f2ff5169fecd496881ba8970f69d5adc977ba4799fedb',
    bitgoKey: 'd4a570c1831c7024cb72533e42690cec9d9150b9aac59001c4a5a3a285518263a9e89f1c709\n' +
        'bcb1e832f2ff5169fecd496881ba8970f69d5adc977ba4799fedb',
    walletPassphrase: '#Bondiola1234',
    walletAddress0: '5Ehp6rRKp4zZxW3zh4SUA79zEj5GDXuNhTy7y7g12DT2kh4C',
    walletAddress1: '5CEJZczUyTF42nSjj8USQ3HFuA9rLbSzJa2t4BgqPN1D22d4',
};
exports.consolidationWrwUser = {
    userKey: '{"iv":"tGqyoisxcojP+QCBKnsBtg==","v":1,"iter":10000,"ks":256,"ts":64,"mode"\n' +
        ':"ccm","adata":"","cipher":"aes","salt":"KKHg339MqDQ=","ct":"cAACEA4x2egsZg\n' +
        'AtS+7igvysPEvNCbk29jnsq+NtzJX/8Zs/CmMkfKtuKKIgqzQA4WufJtAJX0egHYOaIhKpWHwok\n' +
        'oYrA3DSvH+FD4TAuKcMJy5TTaenRbD8rv31CSb4icjaoi0zpImUgfFZcWNdaZ3GrEAKPNh1qgFJ\n' +
        'XwNkpMxVPwic/mv5kqwh+TK0zwRJz5+WDyu8SJ6wT0ACPZ/idifPF5sY4eKuGrnWqbKrleHABKf\n' +
        'wX34m73CsoLCITCrDd9djSEbkf9MMDp9GLzPbxYX/ixyk7Bzrg86ibSj8/+1fcSyUSn4o4O9DZZ\n' +
        'oH9B6IqH5uA1HpOJNhK7SA/HRBqRYuaU4JzJWYFW0fSDY0iCFCvYlzxBytfxGkgu+trnzGjq4pc\n' +
        'YLwvw289PNkDDLS724NcO1rJAdi47ryRUXZKImtnfvb7f5+dpgp+hRFCC4ex3c1W9HEhF85zVgJ\n' +
        'ANFh10a6dZabtSkdoTO2NIBY5oBs0/+ingC2VtwnwhMtpVq4nUr+hvg5ObUR9Ua9eNlQPoPbAxk\n' +
        'SJbW/W3NEMRlN1pNtfdYUNVOdXX/Nce8dacDbuXKKyTlIFAiXXeuecgWVzCbnmEzvmHb6q819wn\n' +
        'crNmsQUt4e4rqNh/tck1EC5g3cPzrUxf5z9dzlupYWbsSwwCAM8mco9Sz6XnDkl5XfKPbAN30Wb\n' +
        'mBUhkYo9mDSYPJgXq/8JD3XaW4I1lO0Afo+Fw3+0HsGBuXUmizD7vZ8FIt9IjCj6l8zk6W2W7ZR\n' +
        'ZYLQxnZHB8ZuoCpaV0fqgE/Ug5WnHeTADmpcf02A8JcZyrwdoaNTEsc2tF36XCYYO3OXdjpj8YM\n' +
        'x4gdsgS9q4NpfJj4g6sNekLA/dqd57NYX7I/94Vj+hIgmwBFIhm1DKfXwLdQR/vKwI5gKPMKXeU\n' +
        '+TYZbEFqfvO9yQ7hQwKi00nQFgFDn75Hday/MkL6XY56d9nuF13J/w1GciA07LV+I1w+WTte6bS\n' +
        'TC66A0XitciWc07MrYYPDlcwu5ZNxVwAGspHJwZ5KhXf2q1qwFPgFYb0rHi32VjSIjuzOIGM3DS\n' +
        '13YoZXYXpZgIL9FJbN1lwrbC4zXzQay7dSYtLODdE8/V6LVteGFrGN1qvxUnTOdkHzGVdYENuFF\n' +
        'RdKOcld8aU2g2QhXFsg8Qd4kYYUnF8Wb3N9RRCQ8v9Lq9IUnWrtq9iTAyuNWxyDHN"}',
    backupKey: '{"iv":"7qW7tFOgvKve47/n5BBHnQ==","v":1,"iter":10000,"ks":256,"ts":64,"mode"\n' +
        ':"ccm","adata":"","cipher":"aes","salt":"ZiJi5j1Q0+w=","ct":"GPSPTslYFHluNl\n' +
        'jGDAWkjJBTPNHa5kVZpoXGgxtkHp6gnWafzFCv9h9ZT/3oSIh1smuzQpsOANn5Gx7Bdvbw8Fwem\n' +
        '+sl4xNd16AwuVBBNNZNyP3QUgcAbKYRcx/ef+aR2D4OWG3nmj1E5LIpG/6FeInigypWtOaLRAHE\n' +
        'euh9YfgvnZopXbZQv1VoN7keqwrVghE1SEILgFTYvIJpWeklMbXEN+Pd8ue9x+8lBgrC7w7b9Yf\n' +
        'Rf+NOaLJ/Z1dlL1nG7jfRlrnzT3ptRLSiZYg5ctn23UowWFOjudA2/XYU+bvGggbnVPNQjLV5eg\n' +
        'kpXXXQzDWvg121IL/MrlSbC1r9hTXv/6BItVxVHCrxLxVRB2vE8D6Fw9w1hwH8mROnq8KaWNYKO\n' +
        'V0pzF070fjD9CXXaYbaOJaG7F5zXCc8D7Lmy5n4Ml4U6BF4OxaBgAOWXkY26R7H48JP2ucID9d9\n' +
        'ofhNmV55I5o4LVXvhV6JytOlQVqqcHjfn0jbVrsxmJiQm5CKaejP5P6GtYz2fp6fpowhcHwUvg4\n' +
        'X6iPFZsRqCupSWFl0wnrUzfXZNNsnH26J6MlkIrOBCbQ5qUCOYaOChoc7yirW8FAQj5jhwxFv1H\n' +
        'hB45MbsuUjIaBhzX2bOcBUfBgfa+jS9Es8pGPloEhDAG6u7PTWnMMqCVtEnsPScyqGvL+Pk5xqg\n' +
        '7ygz9om8t5UCSbmgX1yZiOAI+vuU5uRqk7dbcdWApht8YDFyjNNVnpTcW1RsKhV2SfL+YJkLZZM\n' +
        'PO7QGxjYcdfNgPFghhgAE3/34egmlz6ZUMLE/mas/7116Ui/CbkLPwfguhwShQGy9RY9YmF9nXT\n' +
        'rsSGg3JqXmQH8dpTAZe/aPwZSPKVB//PxvYMuAxP7SPLNcJ8vpHhnxoU9XoyjZasKXAQJTkE0/Y\n' +
        'tv8Z+0tGZowRzuXpIUsVaOIw7O/8SGzynCB6HXcashHbKPtaxRocW855i6ZldWKc1BSXZXZBytT\n' +
        'uyptNLieE1VO44BLAutv+3P0BMqb+Cxu4lXO+zAAwJgh5Psy7QUvD7SEoitC3LJehETmyTmFED3\n' +
        'wO0f4MForhU3QUSyfrqAwkh4dR3b9VtPTPXyONbNcVFIi9PeNMmVbG0uAp2Yy2cgRqD66hM/G3A\n' +
        'RvxPeMHp9z2+KEssgpY206Vo3xwuilQK+4OrkNXXVW+9x+GatimoQsnyvVKraPw=="}',
    bitgoKey: 'cc5b6ca786fefb8339a9ae196f0a76006b6b045e2c4fd9fc1c9422841158ae59c791cbf75bf\n' +
        '488bfeaff75b5f5edd1394c70c14075e7291911990fcb4975baeb',
    walletPassphrase: '#Bondiola1234',
    walletAddress0: '5FxqTKqM9UMsW5rxvKViUxgX6MJjDFXpuoiXHAnr3y4eApps',
    walletAddress1: '5Dc47txC4itx3XEeUBmP15QbSBzHQ2YLJhGTQZssbGBN7VMm',
    walletAddress2: '5CGkKrbTsDcNafwhvqwRpU5grYUBhY8vH1kEB6fQLyeEQuXD',
    walletAddress3: '5FfQ44BKagzpSy8g3MVz2MJijVbxR42CG7Lb8oePiV9cAkfe',
};
exports.westendBlock = {
    blockNumber: 12640277,
    hash: '0x3771101969cc5cf1b37db0bdce589fe9174a1b820341c4d3673ce301352a3d42',
};
