// @bitgo/sdk-coin-tao/dist/test/unit/tao.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const src_1 = require("../../src");
const sinon = __importStar(require("sinon"));
const testData = __importStar(require("./fixtures"));
const resources_1 = require("../resources");
const mocha_1 = require("mocha");
describe('Tao:', function () {
    let bitgo;
    let baseCoin;
    // Test data from wallet 694042b5efbee757e47ec2771cf58a45
    const isWalletAddressTestData = {
        commonKeychain: '6e2235aee215f3909b42bf67c360f5bc6ba7087cbf0ed5ba841dd044ae7c3051722f9be974e8e79fa6c4c93d109dbc618672e36571925df0b5a7c5b015bcd382',
        rootAddress: '5GU55E8X2YVSpd4LApeJR5RsXwQ8AnPdjM37Qz1drLTKh7as',
        receiveAddress: '5EN6LFnhFRtWiwZfFncqBxaNUabTASKnSxoDt2zQvpVX4qVy',
        receiveAddressIndex: 1,
    };
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('tao', src_1.Tao.createInstance);
        bitgo.safeRegister('ttao', src_1.Ttao.createInstance);
        bitgo.initializeTestVars();
        baseCoin = bitgo.coin('ttao');
    });
    describe('isWalletAddress', function () {
        it('should verify root address (index 0)', async function () {
            const keychains = [
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
            ];
            const result = await baseCoin.isWalletAddress({
                address: isWalletAddressTestData.rootAddress,
                keychains,
                index: 0,
            });
            result.should.be.true();
        });
        it('should verify receive address (index > 0)', async function () {
            const keychains = [
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
            ];
            const result = await baseCoin.isWalletAddress({
                address: isWalletAddressTestData.receiveAddress,
                keychains,
                index: isWalletAddressTestData.receiveAddressIndex,
            });
            result.should.be.true();
        });
        it('should throw for address mismatch', async function () {
            const keychains = [
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
            ];
            await baseCoin
                .isWalletAddress({
                address: isWalletAddressTestData.receiveAddress,
                keychains,
                index: 0, // Wrong index for this address
            })
                .should.be.rejectedWith(`address validation failure: ${isWalletAddressTestData.receiveAddress} is not a wallet address`);
        });
        it('should throw for invalid address', async function () {
            const keychains = [
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
                { commonKeychain: isWalletAddressTestData.commonKeychain },
            ];
            await baseCoin
                .isWalletAddress({
                address: 'invalidaddress',
                keychains,
                index: 0,
            })
                .should.be.rejectedWith('invalid address: invalidaddress');
        });
        it('should throw for missing keychains', async function () {
            await baseCoin
                .isWalletAddress({
                address: isWalletAddressTestData.rootAddress,
                keychains: [],
                index: 0,
            })
                .should.be.rejectedWith('missing required param keychains');
        });
    });
    describe.skip('Recover Transactions:', function () {
        const sandBox = sinon.createSandbox();
        const recoveryDestination = '5FJ18ywfrWuRifNyc8aPwQ5ium19Fefwmx18H4XYkDc36F2A';
        const nonce = 2;
        let accountInfoCB;
        let headerInfoCB;
        let getFeeCB;
        beforeEach(function () {
            accountInfoCB = sandBox.stub(src_1.Tao.prototype, 'getAccountInfo');
            headerInfoCB = sandBox.stub(src_1.Tao.prototype, 'getHeaderInfo');
            getFeeCB = sandBox.stub(src_1.Tao.prototype, 'getFee');
            getFeeCB.withArgs(recoveryDestination, testData.wrwUser.walletAddress0, 1510000000000).resolves(15783812856); // TODD: use it later
        });
        (0, mocha_1.afterEach)(function () {
            sandBox.restore();
        });
        it('should generate unsigned sweep correctly', async function () {
            accountInfoCB.withArgs(testData.unsignedSweepUser.walletAddress0).resolves({ nonce: 1, freeBalance: 2989864739 });
            headerInfoCB.resolves({
                headerNumber: testData.westendBlock.blockNumber,
                headerHash: testData.westendBlock.hash,
            });
            const commonKeyChain = 'd4a570c1831c7024cb72533e42690cec9d9150b9aac59001c4a5a3a285518263a9e89f1c709bcb1e832f2ff5169fecd496881ba8970f69d5adc977ba4799fedb'; // TODO: add common key chain
            const unsigned = await baseCoin.recover({ bitgoKey: commonKeyChain, recoveryDestination });
            unsigned.txRequests.should.not.be.undefined();
            unsigned.txRequests.length.should.equal(1);
            unsigned.txRequests[0].transactions.length.should.equal(1);
            unsigned.txRequests[0].walletCoin.should.equal('ttao');
            unsigned.txRequests[0].transactions[0].unsignedTx.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.serializedTx.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.scanIndex.should.equal(0);
            unsigned.txRequests[0].transactions[0].unsignedTx.coin.should.equal('ttao');
            unsigned.txRequests[0].transactions[0].unsignedTx.signableHex.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.derivationPath.should.equal('m/0');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs.length.should.equal(1);
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs[0].address.should.equal('5Ehp6rRKp4zZxW3zh4SUA79zEj5GDXuNhTy7y7g12DT2kh4C');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs[0].valueString.should.equal('2989864739');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs[0].value.should.equal(2989864739);
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs.length.should.equal(1);
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs[0].address.should.equal('5FJ18ywfrWuRifNyc8aPwQ5ium19Fefwmx18H4XYkDc36F2A');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs[0].valueString.should.equal('2989864739');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs[0].coinName.should.equal('ttao');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.type.should.equal('');
            unsigned.txRequests[0].transactions[0].unsignedTx.feeInfo.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.feeInfo.fee.should.equal(0);
            unsigned.txRequests[0].transactions[0].unsignedTx.feeInfo.feeString.should.equal('0');
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.firstValid.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.maxDuration.should.equal(2400);
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.commonKeychain.should.equal('d4a570c1831c7024cb72533e42690cec9d9150b9aac59001c4a5a3a285518263a9e89f1c709bcb1e832f2ff5169fecd496881ba8970f69d5adc977ba4799fedb');
        });
        it('should recover a tx for non-bitgo recoveries', async function () {
            accountInfoCB.withArgs(testData.wrwUser.walletAddress0).resolves({ nonce: 2, freeBalance: 100001853 });
            headerInfoCB.resolves({
                headerNumber: testData.westendBlock.blockNumber,
                headerHash: testData.westendBlock.hash,
            });
            const res = await baseCoin.recover({
                userKey: testData.wrwUser.userKey,
                backupKey: testData.wrwUser.backupKey,
                bitgoKey: testData.wrwUser.bitgoKey,
                walletPassphrase: testData.wrwUser.walletPassphrase,
                recoveryDestination: recoveryDestination,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            res.should.hasOwnProperty('scanIndex');
            // sandBox.assert.calledOnce(baseCoin.getAccountInfo);
            // sandBox.assert.calledOnce(baseCoin.getHeaderInfo);
            const txBuilder = baseCoin.getBuilder().from(res.serializedTx);
            txBuilder
                .validity({
                firstValid: testData.westendBlock.blockNumber,
                maxDuration: baseCoin.SWEEP_TXN_DURATION,
            })
                .referenceBlock(testData.westendBlock.hash);
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.deepEqual(txJson.sender, testData.wrwUser.walletAddress0);
            should_1.default.deepEqual(txJson.blockNumber, testData.westendBlock.blockNumber);
            should_1.default.deepEqual(txJson.referenceBlock, testData.westendBlock.hash);
            should_1.default.deepEqual(txJson.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson.nonce, nonce);
            should_1.default.deepEqual(txJson.tip, 0);
            should_1.default.deepEqual(txJson.transactionVersion, resources_1.txVersion);
            // should.deepEqual(txJson.chainName, chainName);
            should_1.default.deepEqual(txJson.eraPeriod, baseCoin.SWEEP_TXN_DURATION);
        });
        it('should recover a txn for unsigned-sweep recoveries', async function () {
            accountInfoCB.withArgs(testData.unsignedSweepUser.walletAddress0).resolves({ nonce: 1, freeBalance: 2989864739 });
            headerInfoCB.resolves({
                headerNumber: testData.westendBlock.blockNumber,
                headerHash: testData.westendBlock.hash,
            });
            const res = await baseCoin.recover({
                bitgoKey: testData.unsignedSweepUser.bitgoKey,
                recoveryDestination: recoveryDestination,
            });
            res.should.not.be.empty();
            res.txRequests[0].transactions[0].unsignedTx.should.hasOwnProperty('serializedTx');
            res.txRequests[0].transactions[0].unsignedTx.should.hasOwnProperty('scanIndex');
            // sandBox.assert.calledOnce(baseCoin.getAccountInfo);
            // sandBox.assert.calledOnce(baseCoin.getHeaderInfo);
            const txBuilder = baseCoin.getBuilder().from(res.txRequests[0].transactions[0].unsignedTx.serializedTx);
            txBuilder
                .validity({
                firstValid: testData.westendBlock.blockNumber,
                maxDuration: baseCoin.SWEEP_TXN_DURATION,
            })
                .referenceBlock(testData.westendBlock.hash)
                .sender({ address: testData.unsignedSweepUser.walletAddress0 });
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.deepEqual(txJson.sender, testData.unsignedSweepUser.walletAddress0);
            should_1.default.deepEqual(txJson.blockNumber, testData.westendBlock.blockNumber);
            should_1.default.deepEqual(txJson.referenceBlock, testData.westendBlock.hash);
            should_1.default.deepEqual(txJson.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson.nonce, 1);
            should_1.default.deepEqual(txJson.tip, 0);
            should_1.default.deepEqual(txJson.transactionVersion, resources_1.txVersion);
            // should.deepEqual(txJson.chainName, chainName);
            should_1.default.deepEqual(txJson.eraPeriod, baseCoin.SWEEP_TXN_DURATION);
        });
    });
    describe.skip('Build Consolidation Recoveries', function () {
        const sandbox = sinon.createSandbox();
        const baseAddr = testData.consolidationWrwUser.walletAddress0;
        const nonce = 123;
        beforeEach(function () {
            const accountInfoCB = sandbox.stub(src_1.Tao.prototype, 'getAccountInfo');
            accountInfoCB.withArgs(testData.consolidationWrwUser.walletAddress1).resolves({
                nonce: nonce,
                freeBalance: 10000000000,
            });
            accountInfoCB.withArgs(testData.consolidationWrwUser.walletAddress2).resolves({
                nonce: nonce,
                freeBalance: 1510000000000,
            });
            accountInfoCB.withArgs(testData.consolidationWrwUser.walletAddress3).resolves({
                nonce: nonce,
                freeBalance: 1510000000000,
            });
            const headerInfoCB = sandbox.stub(src_1.Tao.prototype, 'getHeaderInfo');
            headerInfoCB.resolves({
                headerNumber: testData.westendBlock.blockNumber,
                headerHash: testData.westendBlock.hash,
            });
            const getFeeCB = sandbox.stub(src_1.Tao.prototype, 'getFee');
            getFeeCB.withArgs(baseAddr, testData.consolidationWrwUser.walletAddress1, 10000000000).resolves(15783812856);
            getFeeCB.withArgs(baseAddr, testData.consolidationWrwUser.walletAddress2, 1510000000000).resolves(15783812856);
            getFeeCB.withArgs(baseAddr, testData.consolidationWrwUser.walletAddress3, 1510000000000).resolves(15783812856);
        });
        (0, mocha_1.afterEach)(function () {
            sandbox.restore();
        });
        it('should build signed consolidation recovery', async function () {
            const res = await baseCoin.recoverConsolidations({
                userKey: testData.consolidationWrwUser.userKey,
                backupKey: testData.consolidationWrwUser.backupKey,
                bitgoKey: testData.consolidationWrwUser.bitgoKey,
                walletPassphrase: testData.consolidationWrwUser.walletPassphrase,
                startingScanIndex: 1,
                endingScanIndex: 4,
            });
            res.should.not.be.empty();
            res.transactions.length.should.equal(2);
            // sandbox.assert.calledThrice(baseCoin.getAccountInfo);
            // sandbox.assert.calledTwice(baseCoin.getHeaderInfo);
            const txn1 = res.transactions[0];
            txn1.should.hasOwnProperty('serializedTx');
            txn1.should.hasOwnProperty('scanIndex');
            txn1.scanIndex.should.equal(2);
            const txBuilder1 = baseCoin.getBuilder().from(txn1.serializedTx);
            txBuilder1
                .validity({
                firstValid: testData.westendBlock.blockNumber,
                maxDuration: baseCoin.MAX_VALIDITY_DURATION,
            })
                .referenceBlock(testData.westendBlock.hash);
            const tx1 = await txBuilder1.build();
            const txJson1 = tx1.toJson();
            should_1.default.deepEqual(txJson1.sender, testData.consolidationWrwUser.walletAddress2);
            should_1.default.deepEqual(txJson1.blockNumber, testData.westendBlock.blockNumber);
            should_1.default.deepEqual(txJson1.referenceBlock, testData.westendBlock.hash);
            should_1.default.deepEqual(txJson1.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson1.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson1.nonce, nonce);
            should_1.default.deepEqual(txJson1.tip, 0);
            should_1.default.deepEqual(txJson1.transactionVersion, resources_1.txVersion);
            // should.deepEqual(txJson1.chainName, chainName);
            // eraPeriod will always round to the next upper power of 2 for any input value, in this case 2400.
            // 4096 is the "highest" value you can set, but the txn still may fail after 2400 blocks.
            const eraPeriod = 4096;
            should_1.default.deepEqual(txJson1.eraPeriod, eraPeriod);
            should_1.default.deepEqual(txJson1.to, baseAddr);
            res.lastScanIndex.should.equal(3);
            const txn2 = res.transactions[1];
            txn2.should.hasOwnProperty('serializedTx');
            txn2.should.hasOwnProperty('scanIndex');
            txn2.scanIndex.should.equal(3);
            const txBuilder2 = baseCoin.getBuilder().from(txn2.serializedTx);
            txBuilder2
                .validity({
                firstValid: testData.westendBlock.blockNumber,
                maxDuration: baseCoin.MAX_VALIDITY_DURATION,
            })
                .referenceBlock(testData.westendBlock.hash);
            const tx2 = await txBuilder2.build();
            const txJson2 = tx2.toJson();
            should_1.default.deepEqual(txJson2.sender, testData.consolidationWrwUser.walletAddress3);
            should_1.default.deepEqual(txJson2.blockNumber, testData.westendBlock.blockNumber);
            should_1.default.deepEqual(txJson2.referenceBlock, testData.westendBlock.hash);
            should_1.default.deepEqual(txJson2.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson2.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson2.nonce, nonce);
            should_1.default.deepEqual(txJson2.tip, 0);
            should_1.default.deepEqual(txJson2.transactionVersion, resources_1.txVersion);
            // should.deepEqual(txJson2.chainName, chainName);
            should_1.default.deepEqual(txJson2.eraPeriod, eraPeriod);
            should_1.default.deepEqual(txJson2.to, baseAddr);
        });
        it('should build unsigned consolidation recoveries', async function () {
            const res = await baseCoin.recoverConsolidations({
                bitgoKey: testData.consolidationWrwUser.bitgoKey,
                startingScanIndex: 1,
                endingScanIndex: 4,
            });
            res.should.not.be.empty();
            res.txRequests.length.should.equal(2);
            sandbox.assert.calledThrice(baseCoin.getAccountInfo);
            sandbox.assert.calledTwice(baseCoin.getHeaderInfo);
            const txn1 = res.txRequests[0].transactions[0].unsignedTx;
            txn1.should.hasOwnProperty('serializedTx');
            txn1.should.hasOwnProperty('signableHex');
            txn1.should.hasOwnProperty('scanIndex');
            txn1.scanIndex.should.equal(2);
            txn1.should.hasOwnProperty('coin');
            txn1.coin.should.equal('ttao');
            txn1.should.hasOwnProperty('derivationPath');
            txn1.derivationPath.should.equal('m/2');
            txn1.should.hasOwnProperty('coinSpecific');
            const coinSpecific1 = txn1.coinSpecific;
            coinSpecific1.should.hasOwnProperty('commonKeychain');
            coinSpecific1.should.hasOwnProperty('firstValid');
            coinSpecific1.firstValid.should.equal(testData.westendBlock.blockNumber);
            coinSpecific1.should.hasOwnProperty('maxDuration');
            coinSpecific1.maxDuration.should.equal(baseCoin.MAX_VALIDITY_DURATION);
            // deserialize the txn and verify the fields are what we expect
            const txBuilder1 = baseCoin.getBuilder().from(txn1.serializedTx);
            // some information isn't deserialized by the from method, so we will
            // supply it again in order to re-build the txn
            txBuilder1
                .validity({
                firstValid: testData.westendBlock.blockNumber,
                maxDuration: baseCoin.MAX_VALIDITY_DURATION,
            })
                .referenceBlock(testData.westendBlock.hash)
                .sender({ address: testData.consolidationWrwUser.walletAddress2 });
            const tx1 = await txBuilder1.build();
            const txJson1 = tx1.toJson();
            should_1.default.deepEqual(txJson1.sender, testData.consolidationWrwUser.walletAddress2);
            should_1.default.deepEqual(txJson1.blockNumber, testData.westendBlock.blockNumber);
            should_1.default.deepEqual(txJson1.referenceBlock, testData.westendBlock.hash);
            should_1.default.deepEqual(txJson1.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson1.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson1.nonce, nonce);
            should_1.default.deepEqual(txJson1.tip, 0);
            should_1.default.deepEqual(txJson1.transactionVersion, resources_1.txVersion);
            // should.deepEqual(txJson1.chainName, chainName);
            // eraPeriod will always round to the next upper power of 2 for any input value, in this case 2400.
            // 4096 is the "highest" value you can set, but the txn still may fail after 2400 blocks.
            const eraPeriod = 4096;
            should_1.default.deepEqual(txJson1.eraPeriod, eraPeriod);
            should_1.default.deepEqual(txJson1.to, baseAddr);
            const txn2 = res.txRequests[1].transactions[0].unsignedTx;
            txn2.should.hasOwnProperty('serializedTx');
            txn2.should.hasOwnProperty('signableHex');
            txn2.should.hasOwnProperty('scanIndex');
            txn2.scanIndex.should.equal(3);
            txn2.should.hasOwnProperty('coin');
            txn2.coin.should.equal('ttao');
            txn2.should.hasOwnProperty('derivationPath');
            txn2.derivationPath.should.equal('m/3');
            txn2.should.hasOwnProperty('coinSpecific');
            const coinSpecific2 = txn2.coinSpecific;
            coinSpecific2.should.hasOwnProperty('commonKeychain');
            coinSpecific2.should.hasOwnProperty('firstValid');
            coinSpecific2.firstValid.should.equal(testData.westendBlock.blockNumber);
            coinSpecific2.should.hasOwnProperty('maxDuration');
            coinSpecific2.maxDuration.should.equal(baseCoin.MAX_VALIDITY_DURATION);
            coinSpecific2.should.hasOwnProperty('commonKeychain');
            coinSpecific2.should.hasOwnProperty('lastScanIndex');
            coinSpecific2.lastScanIndex.should.equal(3);
            // deserialize the txn and verify the fields are what we expect
            const txBuilder2 = baseCoin.getBuilder().from(txn2.serializedTx);
            // some information isn't deserialized by the from method, so we will
            // supply it again in order to re-build the txn
            txBuilder2
                .validity({
                firstValid: testData.westendBlock.blockNumber,
                maxDuration: baseCoin.MAX_VALIDITY_DURATION,
            })
                .referenceBlock(testData.westendBlock.hash)
                .sender({ address: testData.consolidationWrwUser.walletAddress3 });
            const tx2 = await txBuilder2.build();
            const txJson2 = tx2.toJson();
            should_1.default.deepEqual(txJson2.sender, testData.consolidationWrwUser.walletAddress3);
            should_1.default.deepEqual(txJson2.blockNumber, testData.westendBlock.blockNumber);
            should_1.default.deepEqual(txJson2.referenceBlock, testData.westendBlock.hash);
            should_1.default.deepEqual(txJson2.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson2.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson2.nonce, nonce);
            should_1.default.deepEqual(txJson2.tip, 0);
            should_1.default.deepEqual(txJson2.transactionVersion, resources_1.txVersion);
            // should.deepEqual(txJson2.chainName, chainName);
            should_1.default.deepEqual(txJson2.eraPeriod, eraPeriod);
            should_1.default.deepEqual(txJson2.to, baseAddr);
        });
        it('should skip building consolidate transaction if balance is equal to zero', async function () {
            await baseCoin
                .recoverConsolidations({
                userKey: testData.consolidationWrwUser.userKey,
                backupKey: testData.consolidationWrwUser.backupKey,
                bitgoKey: testData.consolidationWrwUser.bitgoKey,
                walletPassphrase: testData.consolidationWrwUser.walletPassphrase,
                startingScanIndex: 1,
                endingScanIndex: 2,
            })
                .should.rejectedWith('Did not find an address with funds to recover');
        });
        it('should throw if startingScanIndex is not ge to 1', async () => {
            await baseCoin
                .recoverConsolidations({
                userKey: testData.consolidationWrwUser.userKey,
                backupKey: testData.consolidationWrwUser.backupKey,
                bitgoKey: testData.consolidationWrwUser.bitgoKey,
                startingScanIndex: -1,
            })
                .should.be.rejectedWith('Invalid starting or ending index to scan for addresses. startingScanIndex: -1, endingScanIndex: 19.');
        });
        it('should throw if scan factor is too high', async () => {
            await baseCoin
                .recoverConsolidations({
                userKey: testData.consolidationWrwUser.userKey,
                backupKey: testData.consolidationWrwUser.backupKey,
                bitgoKey: testData.consolidationWrwUser.bitgoKey,
                startingScanIndex: 1,
                endingScanIndex: 300,
            })
                .should.be.rejectedWith('Invalid starting or ending index to scan for addresses. startingScanIndex: 1, endingScanIndex: 300.');
        });
    });
});
