// @bitgo/abstract-cosmos/dist/src/cosmosCoin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosCoin = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const secp256k1_1 = require("@bitgo/secp256k1");
const bignumber_js_1 = require("bignumber.js");
const buffer_1 = require("buffer");
const crypto_1 = require("crypto");
const _ = __importStar(require("lodash"));
const querystring = __importStar(require("querystring"));
const request = __importStar(require("superagent"));
const url = __importStar(require("url"));
const constants_1 = require("./lib/constants");
const utils_1 = __importDefault(require("./lib/utils"));
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
class CosmosCoin extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new CosmosCoin(bitgo, staticsCoin);
    }
    /**
     * Creates an instance of TransactionBuilderFactory for the coin specific sdk
     */
    getBuilder() {
        throw new Error('Method not implemented.');
    }
    /** @inheritDoc **/
    getBaseFactor() {
        throw new Error('Method not implemented');
    }
    /** @inheritDoc **/
    getChain() {
        return this._staticsCoin.name;
    }
    /** @inheritDoc **/
    getFamily() {
        return this._staticsCoin.family;
    }
    /** @inheritDoc **/
    getFullName() {
        return this._staticsCoin.fullName;
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    /** @inheritDoc **/
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /** @inheritDoc **/
    isValidPub(pub) {
        return utils_1.default.isValidPublicKey(pub);
    }
    /** @inheritDoc **/
    isValidPrv(prv) {
        return utils_1.default.isValidPrivateKey(prv);
    }
    isValidAddress(address) {
        throw new Error('Method not implemented.');
    }
    /**
     * Builds a funds recovery transaction without BitGo
     * @param {RecoveryOptions} params parameters needed to construct and
     * (maybe) sign the transaction
     *
     * @returns {CosmosLikeCoinRecoveryOutput} the serialized transaction hex string and index
     * of the address being swept
     */
    async recover(params) {
        const isUnsignedSweep = this.isUnsignedSweep(params);
        this.validateRecoveryParams(params, isUnsignedSweep);
        const { senderAddress, publicKey, keyShares } = await this.getSenderDetails(params, isUnsignedSweep);
        const [chainId, accountDetails, balances] = await Promise.all([
            this.getChainId(),
            this.getAccountDetails(senderAddress),
            this.getAccountBalance(senderAddress),
        ]);
        const { actualBalance, remainingBalances } = this.processBalances(balances);
        const messages = this.buildTransactionMessages(senderAddress, params.recoveryDestination, actualBalance, remainingBalances);
        return this.buildAndSignTransaction({
            messages,
            chainId,
            accountDetails,
            publicKey: publicKey || '',
            isUnsignedSweep,
            keyShares,
        });
    }
    /**
     * Validates the recovery parameters
     */
    validateRecoveryParams(params, isUnsignedSweep) {
        if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
            throw new Error('invalid recoveryDestination');
        }
        if (!isUnsignedSweep) {
            if (!params.userKey)
                throw new Error('missing userKey');
            if (!params.backupKey)
                throw new Error('missing backupKey');
            if (!params.walletPassphrase)
                throw new Error('missing wallet passphrase');
        }
    }
    /**
     * Checks if this is an unsigned sweep operation
     */
    isUnsignedSweep(params) {
        return !params.userKey && !params.backupKey && !params.walletPassphrase;
    }
    /**
     * Gets sender details including address, public key and key shares
     */
    async getSenderDetails(params, isUnsignedSweep) {
        const MPC = new sdk_core_1.Ecdsa();
        if (isUnsignedSweep) {
            return {
                senderAddress: params.rootAddress,
                publicKey: MPC.deriveUnhardened(params.bitgoKey || '', constants_1.ROOT_PATH).slice(0, 66),
            };
        }
        const keyShares = await this.getKeyShares(params);
        const publicKey = MPC.deriveUnhardened(keyShares.commonKeyChain, constants_1.ROOT_PATH).slice(0, 66);
        return {
            senderAddress: this.getAddressFromPublicKey(publicKey),
            publicKey,
            keyShares,
        };
    }
    /**
     * Gets key shares from recovery parameters
     */
    async getKeyShares(params) {
        if (!params.userKey || !params.backupKey || !params.walletPassphrase) {
            throw new Error('Missing required key parameters');
        }
        const userKey = params.userKey.replace(/\s/g, '');
        const backupKey = params.backupKey.replace(/\s/g, '');
        const walletPassphrase = params.walletPassphrase;
        if (!userKey || !backupKey) {
            throw new Error('Invalid key format');
        }
        return await sdk_core_1.ECDSAUtils.getMpcV2RecoveryKeyShares(userKey, backupKey, walletPassphrase);
    }
    /**
     * Processes account balances and validates sufficient funds
     */
    processBalances(balances) {
        if (!balances?.length) {
            throw new Error('No balance found on account');
        }
        const denomination = this.getDenomination();
        if (!denomination) {
            throw new Error('Invalid denomination');
        }
        let nativeBalance = new bignumber_js_1.BigNumber(0);
        const remainingBalances = [];
        const gasAmountDetails = this.getGasAmountDetails();
        if (!gasAmountDetails?.gasAmount) {
            throw new Error('Invalid gas amount');
        }
        const gasAmount = new bignumber_js_1.BigNumber(gasAmountDetails.gasAmount);
        balances.forEach((balance) => {
            if (!balance.amount) {
                throw new Error('Invalid balance amount');
            }
            if (balance.denom === denomination) {
                nativeBalance = new bignumber_js_1.BigNumber(balance.amount);
            }
            else {
                remainingBalances.push(balance);
            }
        });
        const actualBalance = nativeBalance.minus(gasAmount);
        if (actualBalance.isLessThanOrEqualTo(0)) {
            throw new Error('Did not have enough funds to recover');
        }
        return { nativeBalance, remainingBalances, actualBalance };
    }
    /**
     * Builds transaction messages for all balances
     */
    buildTransactionMessages(senderAddress, recoveryDestination, actualBalance, remainingBalances) {
        const nativeSendMessage = {
            fromAddress: senderAddress,
            toAddress: recoveryDestination,
            amount: [
                {
                    denom: this.getDenomination(),
                    amount: actualBalance.toFixed(),
                },
            ],
        };
        const otherTokenMessages = remainingBalances.map((balance) => ({
            fromAddress: senderAddress,
            toAddress: recoveryDestination,
            amount: [balance],
        }));
        return [...otherTokenMessages, nativeSendMessage];
    }
    /**
     * Builds and signs the transaction
     */
    async buildAndSignTransaction(params) {
        if (!params.chainId) {
            throw new Error('Invalid chain ID');
        }
        const [accountNumber, sequenceNo] = params.accountDetails;
        if (!accountNumber || !sequenceNo) {
            throw new Error('Invalid account details');
        }
        const denomination = this.getDenomination();
        const gasAmountDetails = this.getGasAmountDetails();
        if (!denomination || !gasAmountDetails?.gasAmount || !gasAmountDetails?.gasLimit) {
            throw new Error('Invalid gas configuration');
        }
        const gasBudget = {
            amount: [
                {
                    denom: denomination,
                    amount: gasAmountDetails.gasAmount,
                },
            ],
            gasLimit: gasAmountDetails.gasLimit,
        };
        const txnBuilder = this.getBuilder()
            .getTransferBuilder()
            .messages(params.messages)
            .gasBudget(gasBudget)
            .sequence(Number(sequenceNo))
            .accountNumber(Number(accountNumber))
            .chainId(params.chainId)
            .publicKey(params.publicKey);
        const unsignedTransaction = (await txnBuilder.build());
        if (!unsignedTransaction) {
            throw new Error('Failed to build unsigned transaction');
        }
        if (params.isUnsignedSweep) {
            return {
                signableHex: unsignedTransaction.signablePayload.toString('hex'),
            };
        }
        return this.signTransactionWithMpc(unsignedTransaction, txnBuilder, params.keyShares, params.publicKey);
    }
    /**
     * Signs the transaction with MPC
     */
    /**
     * Signs the transaction using MPC (Multi-Party Computation)
     * @param unsignedTransaction The unsigned transaction to sign
     * @param txnBuilder The transaction builder instance
     * @param keyShares The key shares for MPC signing
     * @param publicKey The public key for verification
     * @returns The signed transaction output
     * @throws Error if validation fails or signing process encounters an error
     */
    async signTransactionWithMpc(unsignedTransaction, txnBuilder, keyShares, publicKey) {
        // Validate inputs
        if (!unsignedTransaction?.signablePayload) {
            throw new Error('Invalid unsigned transaction');
        }
        if (!keyShares?.userKeyShare || !keyShares?.backupKeyShare || !keyShares?.commonKeyChain) {
            throw new Error('Invalid key shares');
        }
        if (!publicKey) {
            throw new Error('Invalid public key');
        }
        try {
            const MPC = new sdk_core_1.Ecdsa();
            const message = unsignedTransaction.signablePayload;
            // Get hash function and compute message hash
            const hashFunction = utils_1.default.getHashFunction() || (0, crypto_1.createHash)('sha256');
            if (!hashFunction) {
                throw new Error('Failed to get hash function');
            }
            const messageHash = hashFunction.update(message).digest();
            // Sign the transaction
            const signature = await sdk_core_1.ECDSAUtils.signRecoveryMpcV2(messageHash, keyShares.userKeyShare, keyShares.backupKeyShare, keyShares.commonKeyChain);
            // Verify the signature
            const signableHex = unsignedTransaction.signablePayload.toString('hex');
            const signableBuffer = buffer_1.Buffer.from(signableHex, 'hex');
            MPC.verify(signableBuffer, signature, this.getHashFunction());
            // Get cosmos key pair and validate
            const cosmosKeyPair = this.getKeyPair(publicKey);
            if (!cosmosKeyPair) {
                throw new Error('Invalid cosmos key pair');
            }
            // Add signature to transaction
            txnBuilder.addSignature({ pub: cosmosKeyPair.getKeys().pub }, buffer_1.Buffer.from(signature.r + signature.s, 'hex'));
            // Build final transaction
            const signedTransaction = await txnBuilder.build();
            if (!signedTransaction) {
                throw new Error('Failed to build signed transaction');
            }
            return {
                serializedTx: signedTransaction.toBroadcastFormat(),
            };
        }
        catch (error) {
            throw new Error(`Failed to sign transaction: ${error.message}`);
        }
    }
    /**
     * Builds a redelegate transaction
     * @param {RecoveryOptions} params parameters needed to construct and
     * (maybe) sign the transaction
     *
     * @returns {CosmosLikeCoinRecoveryOutput} the serialized transaction hex string
     */
    async redelegate(params) {
        if (!params.validatorSrcAddress || !this.isValidAddress(params.validatorSrcAddress)) {
            throw new Error('invalid validatorSrcAddress');
        }
        if (!params.validatorDstAddress || !this.isValidAddress(params.validatorDstAddress)) {
            throw new Error('invalid validatorDstAddress');
        }
        if (!params.userKey) {
            throw new Error('missing userKey');
        }
        if (!params.backupKey) {
            throw new Error('missing backupKey');
        }
        if (!params.walletPassphrase) {
            throw new Error('missing wallet passphrase');
        }
        if (!params.amountToRedelegate) {
            throw new Error('missing amountToRedelegate');
        }
        const userKey = params.userKey.replace(/\s/g, '');
        const backupKey = params.backupKey.replace(/\s/g, '');
        const { userKeyShare, backupKeyShare, commonKeyChain } = await sdk_core_1.ECDSAUtils.getMpcV2RecoveryKeyShares(userKey, backupKey, params.walletPassphrase); // baseAddress is not extracted
        const MPC = new sdk_core_1.Ecdsa();
        const chainId = await this.getChainId();
        const publicKey = MPC.deriveUnhardened(commonKeyChain, constants_1.ROOT_PATH).slice(0, 66);
        const senderAddress = this.getAddressFromPublicKey(publicKey);
        const [accountNumber, sequenceNo] = await this.getAccountDetails(senderAddress);
        const gasBudget = {
            amount: [{ denom: this.getDenomination(), amount: this.getGasAmountDetails().gasAmount }],
            gasLimit: this.getGasAmountDetails().gasLimit,
        };
        const amount = {
            denom: this.getDenomination(),
            amount: new bignumber_js_1.BigNumber(params.amountToRedelegate).toFixed(),
        };
        const sendMessage = [
            {
                delegatorAddress: senderAddress,
                validatorSrcAddress: params.validatorSrcAddress,
                validatorDstAddress: params.validatorDstAddress,
                amount: amount,
            },
        ];
        const txnBuilder = this.getBuilder().getStakingRedelegateBuilder();
        txnBuilder
            .messages(sendMessage)
            .gasBudget(gasBudget)
            .publicKey(publicKey)
            .sequence(Number(sequenceNo))
            .accountNumber(Number(accountNumber))
            .chainId(chainId);
        const unsignedTransaction = (await txnBuilder.build());
        let serializedTx = unsignedTransaction.toBroadcastFormat();
        const signableHex = unsignedTransaction.signablePayload.toString('hex');
        const message = unsignedTransaction.signablePayload;
        const messageHash = (utils_1.default.getHashFunction() || (0, crypto_1.createHash)('sha256')).update(message).digest();
        const signature = await sdk_core_1.ECDSAUtils.signRecoveryMpcV2(messageHash, userKeyShare, backupKeyShare, commonKeyChain);
        const signableBuffer = buffer_1.Buffer.from(signableHex, 'hex');
        MPC.verify(signableBuffer, signature, this.getHashFunction());
        const cosmosKeyPair = this.getKeyPair(publicKey);
        txnBuilder.addSignature({ pub: cosmosKeyPair.getKeys().pub }, buffer_1.Buffer.from(signature.r + signature.s, 'hex'));
        const signedTransaction = await txnBuilder.build();
        serializedTx = signedTransaction.toBroadcastFormat();
        return { serializedTx: serializedTx };
    }
    /** @inheritDoc **/
    async verifyTransaction(params) {
        let totalAmount = new bignumber_js_1.BigNumber(0);
        const { txPrebuild, txParams } = params;
        const rawTx = txPrebuild.txHex;
        if (!rawTx) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const transaction = await this.getBuilder().from(rawTx).build();
        const explainedTx = transaction.explainTransaction();
        if (txParams.recipients && txParams.recipients.length > 0) {
            const filteredRecipients = txParams.recipients?.map((recipient) => _.pick(recipient, ['address', 'amount']));
            const filteredOutputs = explainedTx.outputs.map((output) => _.pick(output, ['address', 'amount']));
            if (!_.isEqual(filteredOutputs, filteredRecipients)) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            // WithdrawDelegatorRewards and ContractCall transaction don't have amount
            if (transaction.type !== sdk_core_1.TransactionType.StakingWithdraw && transaction.type !== sdk_core_1.TransactionType.ContractCall) {
                for (const recipients of txParams.recipients) {
                    totalAmount = totalAmount.plus(recipients.amount);
                }
                if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                    throw new Error('Tx total amount does not match with expected total amount field');
                }
            }
        }
        return true;
    }
    /** @inheritDoc **/
    async explainTransaction(options) {
        if (!options.txHex) {
            throw new Error('missing required txHex parameter');
        }
        try {
            const transactionBuilder = this.getBuilder().from(options.txHex);
            const transaction = await transactionBuilder.build();
            return transaction.explainTransaction();
        }
        catch (e) {
            throw new Error('Invalid transaction: ' + e.message);
        }
    }
    /**
     * Sign a transaction with a single private key
     * @param params parameters in the form of { txPrebuild: {txHex}, prv }
     * @returns signed transaction in the form of { txHex }
     */
    async signTransaction(params) {
        const txHex = params?.txPrebuild?.txHex;
        const privateKey = params?.prv;
        if (!txHex) {
            throw new sdk_core_1.SigningError('missing required txPrebuild parameter: params.txPrebuild.txHex');
        }
        if (!privateKey) {
            throw new sdk_core_1.SigningError('missing required prv parameter: params.prv');
        }
        const txBuilder = this.getBuilder().from(params.txPrebuild.txHex);
        txBuilder.sign({ key: params.prv });
        const transaction = await txBuilder.build();
        if (!transaction) {
            throw new sdk_core_1.SigningError('Failed to build signed transaction');
        }
        const serializedTx = transaction.toBroadcastFormat();
        return {
            txHex: serializedTx,
        };
    }
    /** @inheritDoc **/
    async parseTransaction(params) {
        const transactionExplanation = await this.explainTransaction({ txHex: params.txHex });
        if (!transactionExplanation) {
            throw new Error('Invalid transaction');
        }
        if (transactionExplanation.outputs.length <= 0) {
            return {
                inputs: [],
                outputs: [],
            };
        }
        const senderAddress = transactionExplanation.outputs[0].address;
        const feeAmount = new bignumber_js_1.BigNumber(transactionExplanation.fee.fee === '' ? '0' : transactionExplanation.fee.fee);
        const inputs = [
            {
                address: senderAddress,
                amount: new bignumber_js_1.BigNumber(transactionExplanation.outputAmount).plus(feeAmount).toFixed(),
            },
        ];
        const outputs = transactionExplanation.outputs.map((output) => {
            return {
                address: output.address,
                amount: new bignumber_js_1.BigNumber(output.amount).toFixed(),
            };
        });
        return {
            inputs,
            outputs,
        };
    }
    /**
     * Get the public node url from the Environments constant we have defined
     */
    getPublicNodeUrl() {
        throw new Error('Method not implemented.');
    }
    /**
     * Get account number from public node
     */
    async getAccountFromNode(senderAddress) {
        const nodeUrl = this.getPublicNodeUrl();
        const getAccountPath = '/cosmos/auth/v1beta1/accounts/';
        const fullEndpoint = nodeUrl + getAccountPath + senderAddress;
        try {
            return await request.get(fullEndpoint).send();
        }
        catch (e) {
            console.debug(e);
        }
        throw new Error(`Unable to call endpoint ${getAccountPath + senderAddress} from node: ${nodeUrl}`);
    }
    /**
     * Get balance from public node
     */
    async getBalanceFromNode(senderAddress) {
        const nodeUrl = this.getPublicNodeUrl();
        const getBalancePath = '/cosmos/bank/v1beta1/balances/';
        const fullEndpoint = nodeUrl + getBalancePath + senderAddress;
        try {
            return await request.get(fullEndpoint).send();
        }
        catch (e) {
            console.debug(e);
        }
        throw new Error(`Unable to call endpoint ${getBalancePath + senderAddress} from node: ${nodeUrl}`);
    }
    /**
     * Get chain id from public node
     */
    async getChainIdFromNode() {
        const nodeUrl = this.getPublicNodeUrl();
        const getLatestBlockPath = '/cosmos/base/tendermint/v1beta1/blocks/latest';
        const fullEndpoint = nodeUrl + getLatestBlockPath;
        try {
            return await request.get(fullEndpoint).send();
        }
        catch (e) {
            console.debug(e);
        }
        throw new Error(`Unable to call endpoint ${getLatestBlockPath} from node: ${nodeUrl}`);
    }
    /**
     * Helper to fetch account balance
     */
    async getAccountBalance(senderAddress) {
        const response = await this.getBalanceFromNode(senderAddress);
        if (response.status !== 200) {
            throw new Error('Account not found');
        }
        return response.body.balances;
    }
    /**
     * Helper to fetch chainId
     */
    async getChainId() {
        const response = await this.getChainIdFromNode();
        if (response.status !== 200) {
            throw new Error('Account not found');
        }
        return response.body.block.header.chain_id;
    }
    /**
     * Helper to fetch account number
     */
    async getAccountDetails(senderAddress) {
        const response = await this.getAccountFromNode(senderAddress);
        if (response.status !== 200) {
            throw new Error('Account not found');
        }
        return [response.body.account.account_number, response.body.account.sequence];
    }
    /** @inheritDoc **/
    generateKeyPair(seed) {
        if (!seed) {
            // An extended private key has both a normal 256 bit private key and a 256
            // bit chain code, both of which must be random. 512 bits is therefore the
            // maximum entropy and gives us maximum security against cracking.
            seed = (0, crypto_1.randomBytes)(512 / 8);
        }
        const extendedKey = secp256k1_1.bip32.fromSeed(seed);
        return {
            pub: extendedKey.neutered().toBase58(),
            prv: extendedKey.toBase58(),
        };
    }
    /**
     * Retrieves the address from a public key.
     * @param {string} pubKey - The public key.
     * @returns {string} The corresponding address.
     */
    getAddressFromPublicKey(pubKey) {
        throw new Error('Method not implemented');
    }
    /** @inheritDoc **/
    async isWalletAddress(params) {
        const addressDetails = this.getAddressDetails(params.address);
        if (!this.isValidAddress(addressDetails.address)) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${addressDetails.address}`);
        }
        const rootAddress = params.coinSpecific.rootAddress;
        if (addressDetails.address !== rootAddress) {
            throw new sdk_core_1.UnexpectedAddressError(`address validation failure: ${addressDetails.address} vs ${rootAddress}`);
        }
        return true;
    }
    /** @inheritDoc **/
    getHashFunction() {
        return utils_1.default.getHashFunction();
    }
    /**
     * Process address into address and memo id
     *
     * @param address the address
     * @returns object containing address and memo id
     */
    getAddressDetails(address) {
        const destinationDetails = url.parse(address);
        const destinationAddress = destinationDetails.pathname || '';
        // address doesn't have a memo id
        if (destinationDetails.pathname === address) {
            return {
                address: address,
                memoId: undefined,
            };
        }
        if (!destinationDetails.query) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        const queryDetails = querystring.parse(destinationDetails.query);
        if (!queryDetails.memoId) {
            // if there are more properties, the query details need to contain the memo id property
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        if (Array.isArray(queryDetails.memoId)) {
            throw new sdk_core_1.InvalidAddressError(`memoId may only be given at most once, but found ${queryDetails.memoId.length} instances in address ${address}`);
        }
        if (Array.isArray(queryDetails.memoId) && queryDetails.memoId.length !== 1) {
            // valid addresses can only contain one memo id
            throw new sdk_core_1.InvalidAddressError(`invalid address '${address}', must contain exactly one memoId`);
        }
        const [memoId] = _.castArray(queryDetails.memoId) || undefined;
        if (!this.isValidMemoId(memoId)) {
            throw new sdk_core_1.InvalidMemoIdError(`invalid address: '${address}', memoId is not valid`);
        }
        return {
            address: destinationAddress,
            memoId,
        };
    }
    /**
     * Return boolean indicating whether a memo id is valid
     *
     * @param memoId memo id
     * @returns true if memo id is valid
     */
    isValidMemoId(memoId) {
        return utils_1.default.isValidMemoId(memoId);
    }
    /**
     * Helper method to return the respective coin's base unit
     */
    getDenomination() {
        throw new Error('Method not implemented');
    }
    /**
     * Helper method to fetch gas amount details for respective coin
     */
    getGasAmountDetails() {
        throw new Error('Method not implemented');
    }
    /**
     * Helper method to get key pair for individual coin
     * @param publicKey
     */
    getKeyPair(publicKey) {
        throw new Error('Method not implemented');
    }
    /** @inheritDoc **/
    auditDecryptedKey({ multiSigType, publicKey, prv }) {
        if (multiSigType !== 'tss') {
            throw new Error('Unsupported multiSigType');
        }
        else {
            (0, sdk_lib_mpc_1.auditEcdsaPrivateKey)(prv, publicKey);
        }
    }
}
exports.CosmosCoin = CosmosCoin;
