// @bitgo/abstract-lightning/dist/src/lightning/lightningUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PURPOSE_ALL_OTHERS = exports.PURPOSE_P2TR = exports.PURPOSE_P2WKH = exports.PURPOSE_WRAPPED_P2WKH = exports.lightningNetworkName = exports.signerMacaroonPermissions = void 0;
exports.isLightningCoinName = isLightningCoinName;
exports.getLightningNetwork = getLightningNetwork;
exports.getLightningCoinName = getLightningCoinName;
exports.isValidLightningNetworkName = isValidLightningNetworkName;
exports.isValidLightningNetwork = isValidLightningNetwork;
exports.getStaticsLightningNetwork = getStaticsLightningNetwork;
exports.getUtxolibNetwork = getUtxolibNetwork;
exports.unwrapLightningCoinSpecific = unwrapLightningCoinSpecific;
exports.addIPCaveatToMacaroon = addIPCaveatToMacaroon;
exports.revertXpubPrefix = revertXpubPrefix;
exports.deriveWatchOnlyAccounts = deriveWatchOnlyAccounts;
exports.createWatchOnly = createWatchOnly;
exports.deriveLightningServiceSharedSecret = deriveLightningServiceSharedSecret;
exports.deriveMiddlewareSharedSecret = deriveMiddlewareSharedSecret;
exports.deriveTatSharedSecret = deriveTatSharedSecret;
exports.computeBip32DerivationIndexFromSeed = computeBip32DerivationIndexFromSeed;
const statics = __importStar(require("@bitgo/statics"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const crypto_1 = require("crypto");
const macaroon_1 = require("macaroon");
const bs58check = __importStar(require("bs58check"));
const sdkcore = __importStar(require("@bitgo/sdk-core"));
// https://github.com/lightningnetwork/lnd/blob/master/docs/remote-signing.md#the-signer-node
exports.signerMacaroonPermissions = [
    {
        entity: 'message',
        action: 'write',
    },
    {
        entity: 'signer',
        action: 'generate',
    },
    {
        entity: 'address',
        action: 'read',
    },
    {
        entity: 'onchain',
        action: 'write',
    },
];
exports.lightningNetworkName = ['bitcoin', 'testnet'];
/**
 * Checks if the coin name is a lightning coin name.
 */
function isLightningCoinName(coinName) {
    return coinName === 'lnbtc' || coinName === 'tlnbtc';
}
/**
 * Get the utxolib network for a lightning network.
 */
function getLightningNetwork(networkName) {
    return utxolib.networks[networkName];
}
/**
 * Get the lightning coin name for a utxolib network.
 */
function getLightningCoinName(network) {
    return network === utxolib.networks.bitcoin ? 'lnbtc' : 'tlnbtc';
}
/**
 * Checks if the network name is a valid lightning network name.
 */
function isValidLightningNetworkName(networkName) {
    return exports.lightningNetworkName.includes(networkName);
}
/**
 * Checks if the network is a valid lightning network.
 */
function isValidLightningNetwork(network) {
    return utxolib.isValidNetwork(network) && isValidLightningNetworkName(utxolib.getNetworkName(network));
}
/**
 * Returns the statics network data for a lightning coin.
 */
function getStaticsLightningNetwork(coinName) {
    if (!isLightningCoinName(coinName)) {
        throw new Error(`${coinName} is not a lightning coin`);
    }
    const coin = statics.coins.get(coinName);
    if (!(coin instanceof statics.LightningCoin)) {
        throw new Error('coin is not a lightning coin');
    }
    return coin.network;
}
/**
 * Returns the utxolib network for a lightning coin.
 */
function getUtxolibNetwork(coinName) {
    const networkName = getStaticsLightningNetwork(coinName).utxolibName;
    if (!isValidLightningNetworkName(networkName)) {
        throw new Error('invalid lightning network');
    }
    return getLightningNetwork(networkName);
}
/**
 * Returns coin specific data for a lightning coin.
 */
function unwrapLightningCoinSpecific(obj, coinSpecificPath) {
    if (coinSpecificPath !== 'lnbtc' && coinSpecificPath !== 'tlnbtc') {
        throw new Error(`invalid coinSpecificPath ${coinSpecificPath} for lightning coin`);
    }
    if (coinSpecificPath === 'lnbtc' && 'lnbtc' in obj) {
        return obj.lnbtc;
    }
    if (coinSpecificPath === 'tlnbtc' && 'tlnbtc' in obj) {
        return obj.tlnbtc;
    }
    throw new Error('invalid lightning coin specific');
}
/**
 * Adds an IP caveat to a macaroon and returns the modified macaroon as a Base64 string.
 */
function addIPCaveatToMacaroon(macaroonBase64, ip) {
    const macaroon = (0, macaroon_1.importMacaroon)(macaroonBase64);
    macaroon.addFirstPartyCaveat(`ipaddr ${ip}`);
    return (0, macaroon_1.bytesToBase64)(macaroon.exportBinary());
}
exports.PURPOSE_WRAPPED_P2WKH = 49;
exports.PURPOSE_P2WKH = 84;
exports.PURPOSE_P2TR = 86;
exports.PURPOSE_ALL_OTHERS = 1017;
/**
 * Converts an extended public key (xpub) to the appropriate prefix (ypub, vpub, etc.) based on its purpose and network.
 */
function convertXpubPrefix(xpub, purpose, isMainnet) {
    if (purpose === exports.PURPOSE_P2TR || purpose === exports.PURPOSE_ALL_OTHERS) {
        return xpub;
    }
    const data = bs58check.decode(xpub);
    let versionBytes;
    switch (purpose) {
        case exports.PURPOSE_WRAPPED_P2WKH:
            versionBytes = isMainnet ? Buffer.from([0x04, 0x9d, 0x7c, 0xb2]) : Buffer.from([0x04, 0x4a, 0x52, 0x62]); // ypub/upub for p2sh-p2wpkh
            break;
        case exports.PURPOSE_P2WKH:
            versionBytes = isMainnet ? Buffer.from([0x04, 0xb2, 0x47, 0x46]) : Buffer.from([0x04, 0x5f, 0x1c, 0xf6]); // zpub/vpub for p2wpkh
            break;
        default:
            throw new Error('Unsupported purpose');
    }
    versionBytes.copy(data, 0, 0, 4);
    return bs58check.encode(data);
}
/**
 * Converts a prefix related to purpose and network (ypub, vpub, etc.) to extended public key (xpub).
 */
function revertXpubPrefix(xpub, purpose, isMainnet) {
    // If the purpose is P2TR or ALL_OTHERS, the key is already in the standard format (xpub/tpub),
    // so we return it unmodified. This is the same bypass condition.
    if (purpose === exports.PURPOSE_P2TR || purpose === exports.PURPOSE_ALL_OTHERS) {
        return xpub;
    }
    // 1. Decode the extended public key to get the raw bytes
    const data = bs58check.decode(xpub);
    let versionBytes;
    // 2. Determine the standard prefix (xpub/tpub) based on the network
    switch (purpose) {
        case exports.PURPOSE_WRAPPED_P2WKH:
        case exports.PURPOSE_P2WKH:
            // All standard, non-SegWit-specific keys use the same prefix (xpub for mainnet, tpub for testnet)
            versionBytes = isMainnet
                ? Buffer.from([0x04, 0x88, 0xb2, 0x1e]) // xpub
                : Buffer.from([0x04, 0x35, 0x87, 0xcf]); // tpub
            break;
        default:
            // This case should ideally not be hit if the input key is one of the converted types
            throw new Error('Unsupported purpose for reversal');
    }
    // 3. Overwrite the existing prefix (ypub, zpub, upub, or vpub)
    // with the standard xpub or tpub prefix.
    versionBytes.copy(data, 0, 0, 4);
    // 4. Encode the modified byte data back to a Base58Check string
    return bs58check.encode(data);
}
/**
 * Derives watch-only accounts from the master HD node for the given purposes and network.
 */
function deriveWatchOnlyAccounts(masterHDNode, isMainnet, params = { onlyAddressCreationAccounts: false }) {
    // https://github.com/lightningnetwork/lnd/blob/master/docs/remote-signing.md#required-accounts
    if (masterHDNode.isNeutered()) {
        throw new Error('masterHDNode must not be neutered');
    }
    const purposes = params.onlyAddressCreationAccounts
        ? [exports.PURPOSE_WRAPPED_P2WKH, exports.PURPOSE_P2WKH, exports.PURPOSE_P2TR]
        : [exports.PURPOSE_WRAPPED_P2WKH, exports.PURPOSE_P2WKH, exports.PURPOSE_P2TR, exports.PURPOSE_ALL_OTHERS];
    return purposes.flatMap((purpose) => {
        const maxAccount = purpose === exports.PURPOSE_ALL_OTHERS ? 255 : 0;
        const coinType = purpose !== exports.PURPOSE_ALL_OTHERS || isMainnet ? 0 : 1;
        return Array.from({ length: maxAccount + 1 }, (_, account) => {
            const path = `m/${purpose}'/${coinType}'/${account}'`;
            const derivedNode = masterHDNode.derivePath(path);
            // Ensure the node is neutered (i.e., converted to public key only)
            const neuteredNode = derivedNode.neutered();
            const xpub = convertXpubPrefix(neuteredNode.toBase58(), purpose, isMainnet);
            return {
                purpose,
                coin_type: coinType,
                account,
                xpub,
            };
        });
    });
}
/**
 * Creates a watch-only wallet init data from the provided signer root key and network.
 */
function createWatchOnly(signerRootKey, network) {
    const masterHDNode = utxolib.bip32.fromBase58(signerRootKey, network);
    const getCurrentUnixTimestamp = () => {
        return Math.floor(Date.now() / 1000);
    };
    const master_key_birthday_timestamp = getCurrentUnixTimestamp().toString();
    const master_key_fingerprint = masterHDNode.fingerprint.toString('hex');
    const accounts = deriveWatchOnlyAccounts(masterHDNode, utxolib.isMainnet(network));
    return { master_key_birthday_timestamp, master_key_fingerprint, accounts };
}
/**
 * Derives the shared Elliptic Curve Diffie-Hellman (ECDH) secret between the user's auth extended private key
 * and the Lightning service's public key for secure communication.
 */
function deriveLightningServiceSharedSecret(coinName, userAuthXprv) {
    const publicKey = Buffer.from(getStaticsLightningNetwork(coinName).lightningServicePubKey, 'hex');
    const userAuthHdNode = utxolib.bip32.fromBase58(userAuthXprv);
    return sdkcore.getSharedSecret(userAuthHdNode, publicKey);
}
/**
 * Derives the shared secret for the middleware using a private key and the middleware's public key.
 */
function deriveMiddlewareSharedSecret(coinName, xprv) {
    const publicKey = Buffer.from(getStaticsLightningNetwork(coinName).middlewarePubKey, 'hex');
    const userAuthHdNode = utxolib.bip32.fromBase58(xprv);
    return sdkcore.getSharedSecret(userAuthHdNode, publicKey);
}
/**
 * Derives the shared secret for TAT service using ta private key and the TAT public key.
 */
function deriveTatSharedSecret(coinName, xprv) {
    const publicKey = Buffer.from(getStaticsLightningNetwork(coinName).tatPubKey, 'hex');
    const userAuthHdNode = utxolib.bip32.fromBase58(xprv);
    return sdkcore.getSharedSecret(userAuthHdNode, publicKey);
}
/**
 * Given a seed, compute a BIP32 derivation index.
 * 0 <= index < 2147483648 (largest 31 bit number). This needs to be 2^31 - 1 so that the bip32 library
 * can derive the hardened key.
 * @param seed (optional) If nothing provided, we will generate one randomly
 */
function computeBip32DerivationIndexFromSeed(seed) {
    return ((Buffer.from(utxolib.crypto.sha256(Buffer.from(seed ?? (0, crypto_1.randomBytes)(32).toString('hex'), 'utf8'))).readUint32BE(0) %
        Math.pow(2, 31)) -
        1);
}
