// @bitgo/abstract-lightning/dist/src/lightning/parseWithdrawPsbt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePsbtOutputs = parsePsbtOutputs;
exports.verifyChangeAddress = verifyChangeAddress;
exports.validatePsbtForWithdraw = validatePsbtForWithdraw;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const lightningUtils_1 = require("./lightningUtils");
function parseDerivationPath(derivationPath) {
    const pathSegments = derivationPath.split('/');
    const purpose = Number(pathSegments[1].replace(/'/g, ''));
    const change = Number(pathSegments[pathSegments.length - 2]);
    const addressIndex = Number(pathSegments[pathSegments.length - 1]);
    if (purpose !== lightningUtils_1.PURPOSE_WRAPPED_P2WKH && purpose !== lightningUtils_1.PURPOSE_P2WKH && purpose !== lightningUtils_1.PURPOSE_P2TR) {
        throw new Error(`Unsupported purpose in derivation path: ${purpose}`);
    }
    return { purpose, change, addressIndex };
}
function parsePsbtOutputs(psbt, network) {
    const parsedOutputs = [];
    let bip32Derivation;
    for (let i = 0; i < psbt.data.outputs.length; i++) {
        const output = psbt.data.outputs[i];
        const txOutput = psbt.txOutputs[i];
        let address = '';
        const value = txOutput.value;
        let isChange = false;
        if (output.bip32Derivation && output.bip32Derivation.length > 0) {
            isChange = true;
            bip32Derivation = output.bip32Derivation[0];
        }
        if (txOutput.script) {
            address = utxolib.address.fromOutputScript(txOutput.script, network);
        }
        const valueBigInt = BigInt(value);
        parsedOutputs.push({
            address,
            value: valueBigInt,
            change: isChange,
            bip32Derivation,
        });
    }
    return parsedOutputs;
}
function verifyChangeAddress(output, accounts, network) {
    if (!output.bip32Derivation || !output.bip32Derivation.path) {
        throw new Error(`bip32Derivation path not found for change address`);
    }
    // derivation path example: m/84'/0'/0'/1/0
    const { purpose, change, addressIndex } = parseDerivationPath(output.bip32Derivation.path);
    // Find the corresponding account using the purpose
    const account = accounts.find((acc) => acc.purpose === purpose);
    if (!account) {
        throw new Error(`Account not found for purpose: ${purpose}`);
    }
    // convert upub, vpub, etc prefixes to xpub as utxolib doesn't support these
    const convertedXpub = (0, lightningUtils_1.revertXpubPrefix)(account.xpub, purpose, (0, utxo_lib_1.isMainnet)(network));
    // Create a BIP32 node from the xpub
    const xpubNode = utxolib.bip32.fromBase58(convertedXpub, network);
    // Derive the public key from the xpub using the change and address index
    const derivedPubkey = xpubNode.derive(change).derive(addressIndex).publicKey;
    if (derivedPubkey.toString('hex') !== output.bip32Derivation.pubkey.toString('hex')) {
        throw new Error(`Derived pubkey does not match for address: ${output.address}, derived: ${derivedPubkey.toString('hex')}, expected: ${output.bip32Derivation.pubkey.toString('hex')}`);
    }
    // Determine the correct payment type based on the purpose
    let derivedAddress;
    switch (purpose) {
        case 49: // P2SH-P2WPKH (Nested SegWit)
            derivedAddress = utxolib.payments.p2sh({
                redeem: utxolib.payments.p2wpkh({
                    pubkey: derivedPubkey,
                    network,
                }),
                network,
            }).address;
            break;
        case 84: // P2WPKH (Native SegWit)
            derivedAddress = utxolib.payments.p2wpkh({
                pubkey: derivedPubkey,
                network,
            }).address;
            break;
        case 86: // P2TR (Taproot)
            // P2TR requires x-only pubkey (32 bytes)
            const xOnlyPubkey = derivedPubkey.length === 33 ? derivedPubkey.subarray(1, 33) : derivedPubkey;
            derivedAddress = utxolib.payments.p2tr({
                pubkey: xOnlyPubkey,
                network,
            }, { eccLib: utxolib.ecc }).address;
            break;
        default:
            throw new Error(`Unsupported purpose: ${purpose}`);
    }
    if (derivedAddress !== output.address) {
        throw new Error(`invalid change address: expected ${derivedAddress}, got ${output.address}`);
    }
}
/**
 * Validates the funded psbt before creating the signatures for withdraw.
 */
function validatePsbtForWithdraw(psbtHex, network, recipients, accounts) {
    const parsedPsbt = utxo_lib_1.Psbt.fromHex(psbtHex, { network: network });
    const outputs = parsePsbtOutputs(parsedPsbt, network);
    outputs.forEach((output) => {
        if (output.change) {
            try {
                verifyChangeAddress(output, accounts, network);
            }
            catch (e) {
                throw new Error(`Unable to verify change address: ${e}`);
            }
        }
        else {
            let match = false;
            recipients.forEach((recipient) => {
                if (recipient.address === output.address && BigInt(recipient.amountSat) === output.value) {
                    match = true;
                }
            });
            if (!match) {
                throw new Error(`PSBT output ${output.address} with value ${output.value} does not match any recipient`);
            }
        }
    });
}
