// @bitgo/wasm-utxo/dist/esm/js/descriptorWallet/psbt/parse.js
/**
 * PSBT parsing for descriptor wallets.
 * Moved from @bitgo/utxo-core.
 *
 * This version uses pure TypeScript with PSBT introspection primitives.
 */
import { Psbt } from "../../index.js";
import { fromOutputScriptWithCoin } from "../../address.js";
import { findDescriptorForInput, findDescriptorForOutput, } from "./findDescriptors.js";
import { getVirtualSize } from "../VirtualSize.js";
/**
 * Try to get an address from a script, returning null for non-standard outputs.
 */
function tryGetAddress(script, coin) {
    try {
        return fromOutputScriptWithCoin(script, coin);
    }
    catch {
        return null;
    }
}
/**
 * Parse a PSBT and extract descriptor information.
 *
 * This function uses PSBT introspection to match inputs/outputs against
 * the provided descriptor map.
 *
 * The returned descriptors are reference-identical to those in the input
 * descriptorMap, allowing for `===` comparison.
 *
 * @param psbt - The wasm-utxo Psbt to parse
 * @param descriptorMap - Map of descriptor names to descriptors
 * @param coin - The coin name for address conversion (e.g., "btc", "tbtc")
 * @returns Parsed transaction information
 */
export function parse(psbt, descriptorMap, coin) {
    const rawInputs = psbt.getInputs();
    const rawOutputs = psbt.getOutputs();
    let totalInputValue = 0n;
    let totalOutputValue = 0n;
    let spendAmount = 0n;
    const inputs = rawInputs.map((inputData, i) => {
        const witnessUtxo = inputData.witnessUtxo;
        if (!witnessUtxo) {
            throw new Error(`Missing witnessUtxo for input ${i}`);
        }
        const scriptIdResult = findDescriptorForInput(inputData, descriptorMap);
        if (!scriptIdResult) {
            throw new Error(`No descriptor found for input ${i}`);
        }
        const scriptId = {
            descriptor: scriptIdResult.descriptor,
            index: scriptIdResult.index,
        };
        totalInputValue += witnessUtxo.value;
        return {
            address: fromOutputScriptWithCoin(witnessUtxo.script, coin),
            value: witnessUtxo.value,
            scriptId,
        };
    });
    const outputs = rawOutputs.map((outputData) => {
        const scriptIdResult = findDescriptorForOutput(outputData.script, outputData, descriptorMap);
        const scriptId = scriptIdResult
            ? {
                descriptor: scriptIdResult.descriptor,
                index: scriptIdResult.index,
            }
            : undefined;
        totalOutputValue += outputData.value;
        // Outputs without a matching descriptor are external (spend) outputs
        if (!scriptId) {
            spendAmount += outputData.value;
        }
        return {
            address: tryGetAddress(outputData.script, coin),
            script: outputData.script,
            value: outputData.value,
            scriptId,
        };
    });
    const minerFee = totalInputValue - totalOutputValue;
    // Calculate virtual size using the descriptors from parsed inputs
    const virtualSize = getVirtualSize({
        inputs: inputs.map((input) => input.scriptId.descriptor),
        outputs: outputs.map((output) => ({ script: output.script })),
    });
    return {
        inputs,
        outputs,
        spendAmount,
        minerFee,
        virtualSize,
    };
}
/**
 * Parse a serialized PSBT buffer with descriptor information.
 *
 * This is a convenience function that creates a Psbt from bytes before parsing.
 *
 * @param psbtBytes - The serialized PSBT bytes
 * @param descriptorMap - Map of descriptor names to descriptors
 * @param coin - The coin name for address conversion
 * @returns Parsed transaction information
 */
export function parseFromBytes(psbtBytes, descriptorMap, coin) {
    const psbt = Psbt.deserialize(psbtBytes);
    return parse(psbt, descriptorMap, coin);
}
