// @bitgo/sdk-core/dist/src/bitgo/wallet/wallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wallet = exports.ManageUnspentsOptions = void 0;
/**
 * @prettier
 */
const public_types_1 = require("@bitgo/public-types");
const statics_1 = require("@bitgo/statics");
const assert_1 = __importDefault(require("assert"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const t = __importStar(require("io-ts"));
const _ = __importStar(require("lodash"));
const common = __importStar(require("../../common"));
const address_book_1 = require("../address-book");
const bitcoin_1 = require("../bitcoin");
const ecdh_1 = require("../ecdh");
const errors_1 = require("../errors");
const internal_1 = require("../internal");
const internal = __importStar(require("../internal/internal"));
const keychain_1 = require("../keychain");
const lightningWalletUtil_1 = require("../lightning/lightningWalletUtil");
const pendingApproval_1 = require("../pendingApproval");
const staking_1 = require("../staking");
const trading_1 = require("../trading");
const tss_1 = require("../tss");
const utils_1 = require("../utils");
const postWithCodec_1 = require("../utils/postWithCodec");
const ecdsa_1 = require("../utils/tss/ecdsa");
const eddsa_1 = __importDefault(require("../utils/tss/eddsa"));
const txRequest_1 = require("../utils/txRequest");
const BuildParams_1 = require("./BuildParams");
const debug = require('debug')('bitgo:v2:wallet');
const whitelistedSendParams = public_types_1.TxSendBody.type.types.flatMap((t) => Object.keys(t.props));
var ManageUnspentsOptions;
(function (ManageUnspentsOptions) {
    ManageUnspentsOptions[ManageUnspentsOptions["BUILD_ONLY"] = 0] = "BUILD_ONLY";
    ManageUnspentsOptions[ManageUnspentsOptions["BUILD_SIGN_SEND"] = 1] = "BUILD_SIGN_SEND";
})(ManageUnspentsOptions || (exports.ManageUnspentsOptions = ManageUnspentsOptions = {}));
function isPrebuildTransactionResult(prebuildTx) {
    if (!prebuildTx || typeof prebuildTx === 'string') {
        return false;
    }
    return prebuildTx.walletId !== undefined;
}
class Wallet {
    constructor(bitgo, baseCoin, walletData) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
        this._wallet = walletData;
        const userId = _.get(bitgo, '_user.id');
        if (_.isString(userId)) {
            const userDetails = _.find(walletData.users, { user: userId });
            this._permissions = _.get(userDetails, 'permissions');
        }
        if (baseCoin?.supportsTss() && this._wallet.multisigType === 'tss') {
            switch (baseCoin.getMPCAlgorithm()) {
                case 'ecdsa':
                    if (walletData.multisigTypeVersion === 'MPCv2') {
                        this.tssUtils = new ecdsa_1.EcdsaMPCv2Utils(bitgo, baseCoin, this);
                    }
                    else {
                        this.tssUtils = new ecdsa_1.EcdsaUtils(bitgo, baseCoin, this);
                    }
                    break;
                case 'eddsa':
                    this.tssUtils = new eddsa_1.default(bitgo, baseCoin, this);
                    break;
                default:
                    this.tssUtils = undefined;
            }
        }
    }
    /**
     * Build a URL using this wallet's id which can be used for BitGo API operations
     * @param extra API specific string to append to the wallet id
     */
    url(extra = '') {
        return this.baseCoin.url('/wallet/' + this.id() + extra);
    }
    /**
     * Get this wallet's id
     */
    id() {
        return this._wallet.id;
    }
    /**
     * Get the number of approvals required for spending funds from this wallet
     */
    approvalsRequired() {
        return this._wallet.approvalsRequired;
    }
    /**
     * Get the current balance of this wallet
     */
    balance() {
        return this._wallet.balance;
    }
    /** @deprecated use codec instead: t.exact(BuildParams).encode(v) */
    prebuildWhitelistedParams() {
        return BuildParams_1.buildParamKeys;
    }
    /**
     * This is a strict sub-set of prebuildWhitelistedParams
     */
    prebuildConsolidateAccountParams() {
        return [
            'consolidateAddresses',
            'nftCollectionId',
            'nftId',
            'feeRate',
            'maxFeeRate',
            'memo',
            'validFromBlock',
            'validToBlock',
            'preview',
            'keepAlive',
            'apiVersion',
        ];
    }
    /**
     * Get the confirmed balance of this wallet
     */
    confirmedBalance() {
        return this._wallet.confirmedBalance;
    }
    /**
     * Get the spendable balance of this wallet
     */
    spendableBalance() {
        return this._wallet.spendableBalance;
    }
    /**
     * Get a string representation of the balance of this wallet
     *
     * This is useful when balances have the potential to overflow standard javascript numbers
     */
    balanceString() {
        return this._wallet.balanceString;
    }
    /**
     * Get a string representation of the confirmed balance of this wallet
     *
     * This is useful when balances have the potential to overflow standard javascript numbers
     */
    confirmedBalanceString() {
        return this._wallet.confirmedBalanceString;
    }
    /**
     * Get a string representation of the spendable balance of this wallet
     *
     * This is useful when balances have the potential to overflow standard javascript numbers
     */
    spendableBalanceString() {
        return this._wallet.spendableBalanceString;
    }
    /**
     * Get the coin identifier for the type of coin this wallet holds
     */
    coin() {
        return this._wallet.coin;
    }
    type() {
        return this._wallet.type || 'hot';
    }
    multisigType() {
        return this._wallet.multisigType;
    }
    multisigTypeVersion() {
        return this._wallet.multisigTypeVersion;
    }
    subType() {
        return this._wallet.subType;
    }
    /**
     * Get the label (name) for this wallet
     */
    label() {
        return this._wallet.label;
    }
    flags() {
        return this._wallet.walletFlags ?? [];
    }
    flag(name) {
        return this.flags().find((flag) => flag.name === name)?.value;
    }
    /**
     * Get the public object ids for the keychains on this wallet.
     */
    keyIds() {
        return this._wallet.keys;
    }
    /**
     * Get a receive address for this wallet
     */
    receiveAddress() {
        return this._wallet.receiveAddress?.address;
    }
    /**
     * Get the wallet id of the wallet that this wallet was migrated from.
     *
     * For example, if this is a BCH wallet that was created from a BTC wallet,
     * the BCH wallet migrated from field would have the BTC wallet id.
     */
    migratedFrom() {
        return this._wallet.migratedFrom;
    }
    /**
     * Return the token flush thresholds for this wallet
     * @return {*|Object} pairs of { [tokenName]: thresholds } base units
     */
    tokenFlushThresholds() {
        if (this.baseCoin.getFamily() !== 'eth') {
            throw new Error('not supported for this wallet');
        }
        return this._wallet.coinSpecific.tokenFlushThresholds;
    }
    /**
     * Get wallet properties which are specific to certain coin implementations
     */
    coinSpecific() {
        return this._wallet.coinSpecific;
    }
    /**
     * Get all pending approvals on this wallet
     */
    pendingApprovals() {
        return this._wallet.pendingApprovals.map((currentApproval) => {
            return new pendingApproval_1.PendingApproval(this.bitgo, this.baseCoin, currentApproval, this);
        });
    }
    /**
     * Refresh the wallet object by syncing with the back-end
     * @param params
     * @returns {Wallet}
     */
    async refresh(params = {}) {
        this._wallet = await this.bitgo.get(this.url()).result();
        return this;
    }
    /**
     * List the transactions for a given wallet
     * @param params
     * @returns {*}
     */
    async transactions(params = {}) {
        const query = {};
        if (params.prevId) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            query.prevId = params.prevId;
        }
        if (params.limit) {
            if (!_.isNumber(params.limit)) {
                throw new Error('invalid limit argument, expecting number');
            }
            query.limit = params.limit;
        }
        return await this.bitgo
            .get(this.baseCoin.url('/wallet/' + this._wallet.id + '/tx'))
            .query(query)
            .result();
    }
    /**
     * Return a list of nft tokens for this wallet. Will always return undefined if the wallet
     * was not initialized with the allTokens flag.
     *
     * @returns {NftBalance[] | undefined}
     */
    nftBalances() {
        if (this._wallet.nfts) {
            return Object.values(this._wallet.nfts).map((nftData) => nftData);
        }
        return undefined;
    }
    /**
     * Return a list of unsupported nft tokens for this wallet. Will always return undefined if the wallet
     * was not initialized with the allTokens flag.
     *
     * @returns {NftBalance[] | undefined}
     */
    unsupportedNftBalances() {
        if (this._wallet.unsupportedNfts) {
            return Object.values(this._wallet.unsupportedNfts).map((nftData) => nftData);
        }
        return undefined;
    }
    /**
     * Returns a list of the wallets nft & unsupported nfts.
     *
     * @returns {NftBalance[]}
     */
    async getNftBalances() {
        const walletData = await this.bitgo.get(this.url()).query({ allTokens: true }).result();
        const supportedNfts = walletData?.nfts ? Object.values(walletData.nfts).map((balance) => balance) : [];
        const unsupportedNfts = walletData?.unsupportedNfts
            ? Object.values(walletData.unsupportedNfts).map((balance) => balance)
            : [];
        return [...supportedNfts, ...unsupportedNfts];
    }
    /**
     * List the transactions for a given wallet
     * @param params
     *  - txHash the transaction hash to search for
     * @returns {*}
     */
    async getTransaction(params = {}) {
        common.validateParams(params, ['txHash'], []);
        const paginatedOptions = {};
        if (!_.isUndefined(params.prevId)) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            paginatedOptions.prevId = params.prevId;
        }
        if (!_.isUndefined(params.limit)) {
            if (!_.isInteger(params.limit) || params.limit < 1) {
                throw new Error('invalid limit argument, expecting positive integer');
            }
            paginatedOptions.limit = params.limit;
        }
        const query = paginatedOptions;
        if (params.includeRbf) {
            query['includeRbf'] = params.includeRbf;
        }
        return await this.bitgo
            .get(this.url('/tx/' + params.txHash))
            .query(query)
            .result();
    }
    /**
     * List the transfers for a given wallet
     * @param params
     * @returns {*}
     */
    async transfers(params = {}) {
        const query = {};
        if (params.prevId) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            query.prevId = params.prevId;
        }
        if (params.limit) {
            if (!_.isNumber(params.limit)) {
                throw new Error('invalid limit argument, expecting number');
            }
            query.limit = params.limit;
        }
        if (params.allTokens) {
            if (!_.isBoolean(params.allTokens)) {
                throw new Error('invalid allTokens argument, expecting boolean');
            }
            query.allTokens = params.allTokens;
        }
        if (params.searchLabel) {
            if (!_.isString(params.searchLabel)) {
                throw new Error('invalid searchLabel argument, expecting string');
            }
            query.searchLabel = params.searchLabel;
        }
        if (params.address) {
            if (!_.isArray(params.address) && !_.isString(params.address)) {
                throw new Error('invalid address argument, expecting string or array');
            }
            if (_.isArray(params.address)) {
                params.address.forEach((address) => {
                    if (!_.isString(address)) {
                        throw new Error('invalid address argument, expecting array of address strings');
                    }
                });
            }
            query.address = params.address;
        }
        if (params.dateGte) {
            if (!_.isString(params.dateGte)) {
                throw new Error('invalid dateGte argument, expecting string');
            }
            query.dateGte = params.dateGte;
        }
        if (params.dateLt) {
            if (!_.isString(params.dateLt)) {
                throw new Error('invalid dateLt argument, expecting string');
            }
            query.dateLt = params.dateLt;
        }
        if (!_.isNil(params.valueGte)) {
            if (!_.isNumber(params.valueGte)) {
                throw new Error('invalid valueGte argument, expecting number');
            }
            query.valueGte = params.valueGte;
        }
        if (!_.isNil(params.valueLt)) {
            if (!_.isNumber(params.valueLt)) {
                throw new Error('invalid valueLt argument, expecting number');
            }
            query.valueLt = params.valueLt;
        }
        if (!_.isNil(params.includeHex)) {
            if (!_.isBoolean(params.includeHex)) {
                throw new Error('invalid includeHex argument, expecting boolean');
            }
            query.includeHex = params.includeHex;
        }
        if (!_.isNil(params.state)) {
            if (!Array.isArray(params.state) && !_.isString(params.state)) {
                throw new Error('invalid state argument, expecting string or array');
            }
            if (Array.isArray(params.state)) {
                params.state.forEach((state) => {
                    if (!_.isString(state)) {
                        throw new Error('invalid state argument, expecting array of state strings');
                    }
                });
            }
            query.state = params.state;
        }
        if (!_.isNil(params.type)) {
            if (!_.isString(params.type)) {
                throw new Error('invalid type argument, expecting string');
            }
            query.type = params.type;
        }
        if (!_.isNil(params.decorateUtxoSpecificFields)) {
            if (!_.isBoolean(params.decorateUtxoSpecificFields)) {
                throw new Error('invalid includeHex argument, expecting boolean');
            }
            query.decorateUtxoSpecificFields = params.decorateUtxoSpecificFields;
        }
        return await this.bitgo.get(this.url('/transfer')).query(query).result();
    }
    /**
     * Get transfers on this wallet
     * @param params
     */
    async getTransfer(params = {}) {
        common.validateParams(params, ['id'], []);
        return await this.bitgo.get(this.url('/transfer/' + params.id)).result();
    }
    /**
     * Get a transaction by sequence id for a given wallet
     * @param params
     */
    async transferBySequenceId(params = {}) {
        common.validateParams(params, ['sequenceId'], []);
        return await this.bitgo.get(this.url('/transfer/sequenceId/' + params.sequenceId)).result();
    }
    /**
     * Get the maximum amount you can spend in a single transaction
     *
     * @param {Object} params - parameters object
     * @param {Number} params.limit - maximum number of selectable unspents
     * @param {Number | String} params.minValue - the minimum value of unspents to use in satoshis
     * @param {Number | String} params.maxValue - the maximum value of unspents to use in satoshis
     * @param {Number} params.minHeight - the minimum height of unspents on the block chain to use
     * @param {Number} params.minConfirms - all selected unspents will have at least this many confirmations
     * @param {Boolean} params.enforceMinConfirmsForChange - Enforces minConfirms on change inputs
     * @param {Number} params.feeRate - fee rate to use in calculation of maximum spendable in satoshis/kB
     * @param {Number} params.maxFeeRate - upper limit for feeRate in satoshis/kB
     * @param {String} params.recipientAddress - recipient addresses for a more accurate calculation of the maximum available to send
     * @returns {{maximumSpendable: Number, coin: String}}
     * NOTE : feeTxConfirmTarget omitted on purpose because gauging the maximum spendable amount with dynamic fees does not make sense
     */
    async maximumSpendable(params = {}) {
        const filteredParams = _.pick(params, [
            'enforceMinConfirmsForChange',
            'feeRate',
            'limit',
            'maxFeeRate',
            'maxValue',
            'minConfirms',
            'minHeight',
            'minValue',
            'plainTarget',
            'recipientAddress',
            'target',
        ]);
        return await this.bitgo.get(this.url('/maximumSpendable')).query(filteredParams).result();
    }
    /**
     * List the unspents for a given wallet
     * @param params
     * @returns {*}
     */
    async unspents(params = {}) {
        const query = _.pick(params, [
            'chains',
            'limit',
            'maxValue',
            'minConfirms',
            'minHeight',
            'minValue',
            'prevId',
            'segwit',
            'target',
            'unspentIds',
        ]);
        return this.bitgo.get(this.url('/unspents')).query(query).result();
    }
    /**
     * Consolidate or fanout unspents on a wallet
     *
     * @param {String} routeName - either `consolidate` or `fanout`
     *
     * @param {Object} params - parameters object
     *
     * Wallet parameters:
     * @param {String} params.walletPassphrase - the users wallet passphrase
     * @param {String} params.xprv - the private key in string form if the walletPassphrase is not available
     *
     * Fee parameters:
     * @param {Number} params.feeRate - The fee rate to use for the consolidation in satoshis/kB
     * @param {Number} params.maxFeeRate - upper limit for feeRate in satoshis/kB
     * @param {Number} params.maxFeePercentage - the maximum relative portion that you're willing to spend towards fees
     * @param {Number} params.feeTxConfirmTarget - estimate the fees to aim for first confirmation with this number of blocks
     *
     * Input parameters:
     * @param {Number | String} params.minValue - the minimum value of unspents to use in satoshis
     * @param {Number | String} params.maxValue - the maximum value of unspents to use in satoshis
     * @param {Number} params.minHeight - the minimum height of unspents on the block chain to use
     * @param {Number} params.minConfirms - all selected unspents will have at least this many confirmations
     * @param {Boolean} params.enforceMinConfirmsForChange - if true, minConfirms also applies to change outputs
     * @param {Number} params.limit                for routeName === 'consolidate'
     *                 params.maxNumInputsToUse    for routeName === 'fanout'
     *                  - maximum number of unspents you want to use in the transaction
     * Output parameters:
     * @param {Number} params.numUnspentsToMake - the number of new unspents to make
     * @param {Boolean} params.bulk - if set to True, this enables the consolidation of large number of unspents by creating multiple transactions,
     *                                with each transaction composed of 200 unspents, except for the last transaction which may have fewer unspents.
     */
    async manageUnspents(routeName, params = {}, option = ManageUnspentsOptions.BUILD_SIGN_SEND) {
        if (this._wallet.type === 'custodial' && routeName === 'consolidate') {
            return this.initiateTransaction({ ...params, type: 'consolidate' });
        }
        else if (this._wallet.type === 'custodial' && routeName === 'fanout') {
            throw new Error('Fanout is not supported for custodial wallets');
        }
        common.validateParams(params, [], ['walletPassphrase', 'xprv']);
        const reqId = new utils_1.RequestTracer();
        const fanoutInputFormat = params.maxNumInputsToUse ? 'maxNumInputsToUse' : 'unspents';
        const filteredParams = _.pick(params, [
            'feeRate',
            'maxFeeRate',
            'maxFeePercentage',
            'feeTxConfirmTarget',
            'minValue',
            'maxValue',
            'minHeight',
            'minConfirms',
            'enforceMinConfirmsForChange',
            'targetAddress',
            'txFormat',
            'bulk',
            routeName === 'consolidate' ? 'limit' : fanoutInputFormat,
            'numUnspentsToMake',
        ]);
        this.bitgo.setRequestTracer(reqId);
        const buildResponse = await this.bitgo
            .post(this.url(`/${routeName}Unspents`))
            .send(filteredParams)
            .result();
        if (option === ManageUnspentsOptions.BUILD_ONLY) {
            return buildResponse;
        }
        const keychains = (await this.baseCoin
            .keychains()
            .getKeysForSigning({ wallet: this, reqId }));
        // Validate that the platform-built transaction matches user parameters
        const txPrebuilds = Array.isArray(buildResponse) ? buildResponse : [buildResponse];
        await Promise.all(txPrebuilds.map((txPrebuild) => this.baseCoin.verifyTransaction({
            txParams: params,
            txPrebuild,
            wallet: this,
            verification: {
                ...(params.verification ?? {}),
                keychains: {
                    user: keychains[0],
                    backup: keychains[1],
                    bitgo: keychains[2],
                },
            },
            reqId,
        })));
        const transactionParams = {
            ...params,
            keychain: keychains[0],
            pubs: keychains.map((k) => {
                (0, assert_1.default)(k.pub);
                return k.pub;
            }),
            // Building PSBTs with the bulk flag does not include the previous transaction for non-segwit inputs.
            // Manually override the signing and validating to not fail.
            allowNonSegwitSigningWithoutPrevTx: !!params.bulk,
        };
        const selectParams = _.pick(params, ['comment', 'otp', 'bulk']);
        const response = await Promise.all(txPrebuilds.map(async (txPrebuild) => {
            const signedTransaction = await this.signTransaction({
                ...transactionParams,
                txPrebuild,
            });
            const finalTxParams = _.extend({}, signedTransaction, selectParams, {
                type: routeName,
            });
            this.bitgo.setRequestTracer(reqId);
            return this.sendTransaction(finalTxParams, reqId);
        }));
        return Array.isArray(buildResponse) ? response : response[0];
    }
    /**
     * Manage the unspent reservations on the wallet
     *
     * @param params.create - create a new reservation
     * @param params.modify - modify an existing reservation
     * @param params.delete - delete an existing reservation
     */
    async manageUnspentReservations(params) {
        const filteredParams = _.pick(params, ['create', 'modify', 'delete']);
        this.bitgo.setRequestTracer(new utils_1.RequestTracer());
        // The URL cannot contain the coinName, so we remove it from the URL
        const url = this.url(`/reservedunspents`).replace(`/${this.baseCoin.getChain()}`, '');
        if (filteredParams.create) {
            const filteredCreateParams = _.pick(params.create, ['unspentIds', 'expireTime']);
            return this.bitgo.post(url).send(filteredCreateParams).result();
        }
        else if (filteredParams.modify) {
            const filteredModifyParams = _.pick(params.modify, ['unspentIds', 'changes']);
            return this.bitgo.put(url).send(filteredModifyParams).result();
        }
        else if (filteredParams.delete) {
            const filteredDeleteParams = _.pick(params.delete, ['id']);
            return this.bitgo.del(url).query(filteredDeleteParams).result();
        }
        else {
            throw new Error('Did not detect a creation, modification, or deletion request.');
        }
    }
    /**
     * Consolidate unspents on a wallet
     *
     * @param {Object} params - parameters object
     * @param {String} params.walletPassphrase - the users wallet passphrase
     * @param {String} params.xprv - the private key in string form if the walletPassphrase is not available
     * @param {Number} params.feeRate - The fee rate to use for the consolidation in satoshis/kB
     * @param {Number} params.maxFeeRate - upper limit for feeRate in satoshis/kB
     * @param {Number} params.maxFeePercentage - the maximum relative portion that you're willing to spend towards fees
     * @param {Number} params.feeTxConfirmTarget - estimate the fees to aim for first confirmation with this number of blocks
     * @param {Number | String} params.minValue - the minimum value of unspents to use in satoshis
     * @param {Number | String} params.maxValue - the maximum value of unspents to use in satoshis
     * @param {Number} params.minHeight - the minimum height of unspents on the block chain to use
     * @param {Number} params.minConfirms - all selected unspents will have at least this many confirmations
     * @param {Boolean} params.enforceMinConfirmsForChange - if true, minConfirms also applies to change outputs
     * @param {Number} params.limit                for routeName === 'consolidate'
     *                 params.maxNumInputsToUse    for routeName === 'fanout'
     *                  - maximum number of unspents you want to use in the transaction
     * @param {Number} params.numUnspentsToMake - the number of new unspents to make. It is not applicable for if bulk consolidate.
     * @param {Boolean} params.bulk - if set to True, this enables the consolidation of large number of unspents by creating multiple transactions,
     *                                with each transaction composed of 200 unspents, except for the last transaction which may have fewer unspents.
     */
    async consolidateUnspents(params = {}, option = ManageUnspentsOptions.BUILD_SIGN_SEND) {
        return this.manageUnspents('consolidate', params, option);
    }
    /**
     * Fanout unspents on a wallet
     *
     * @param {Object} params - parameters object
     * @param {String} params.walletPassphrase - the users wallet passphrase
     * @param {String} params.xprv - the private key in string form if the walletPassphrase is not available
     * @param {Number | String} params.minValue - the minimum value of unspents to use
     * @param {Number | String} params.maxValue - the maximum value of unspents to use
     * @param {Number} params.minHeight - the minimum height of unspents on the block chain to use
     * @param {Number} params.minConfirms - all selected unspents will have at least this many confirmations
     * @param {Number} params.maxFeePercentage - the maximum proportion of an unspent you are willing to lose to fees
     * @param {Number} params.feeTxConfirmTarget - estimate the fees to aim for first confirmation with this number of blocks
     * @param {Number} params.feeRate - The desired fee rate for the transaction in satoshis/kB
     * @param {Number} params.maxFeeRate - The max limit for a fee rate in satoshis/kB
     * @param {Number} params.maxNumInputsToUse - the number of unspents you want to use in the transaction
     * @param {Number} params.numUnspentsToMake - the number of new unspents to make
     *
     * @param {ManageUnspentsOptions} option - flag to toggle build and send or build only
     */
    async fanoutUnspents(params = {}, option = ManageUnspentsOptions.BUILD_SIGN_SEND) {
        return this.manageUnspents('fanout', params, option);
    }
    /**
     * Set the token flush thresholds for the wallet. Updates the wallet.
     * Tokens will only be flushed from forwarder contracts if the balance is greater than the threshold defined here.
     * @param thresholds {Object} - pairs of { [tokenName]: threshold } (base units)
     */
    async updateTokenFlushThresholds(thresholds = {}) {
        if (this.baseCoin.getFamily() !== 'eth') {
            throw new Error('not supported for this wallet');
        }
        this._wallet = await this.bitgo
            .put(this.url())
            .send({
            tokenFlushThresholds: thresholds,
        })
            .result();
    }
    /**
        * Updates the wallet. Sets flags for deployForwardersManually and flushForwardersManually of the wallet.
        * @param forwarderFlags {Object} - {
          "coinSpecific": {
            [coinName]: {
              "deployForwardersManually": {Boolean},
              "flushForwardersManually": {Boolean}
            }
          }
        }
        */
    async updateForwarders(forwarderFlags = {}) {
        if (this.baseCoin.getFamily() !== 'eth') {
            throw new Error('not supported for this wallet');
        }
        this._wallet = await this.bitgo.put(this.url()).send(forwarderFlags).result();
    }
    /**
     * To manually deploy an ETH address
     *
     * @param {Object} params - parameters object
     * @param {String} [params.address] - addressId
     * @param {String} [params.id] - addressId could be received also as id
     * @returns {Object} Http response
     */
    async deployForwarders(params) {
        if (_.isUndefined(params.address) && _.isUndefined(params.id)) {
            throw new Error('address or id of address required');
        }
        let query;
        if (params.address) {
            query = params.address;
        }
        else {
            query = params.id;
        }
        const url = this.url(`/address/${encodeURIComponent(query)}/deployment`);
        this._wallet = await this.bitgo.post(url).send(params).result();
        return this._wallet;
    }
    /**
     * To manually forward tokens from an ETH or CELO address
     *
     * @param {Object} params - parameters object
     * @param {String} params.tokenName - Name of token that needs to be forwarded from the address
     * @param {String} [params.address] -
     * @param {String} [params.address] - addressId
     * @param {String} [params.id] - addressId could be received also as id
     * @param {String} [params.gasPrice] - Explicit gas price to use when forwarding token from the forwarder contract (ETH and Celo only). If not given, defaults to the current estimated network gas price.
     * @param {String} [params.eip1559] - Specify eip1559 fee parameters in token forwarding transaction.
     * @returns {Object} Http response
     */
    async flushForwarderToken(params) {
        if (_.isUndefined(params.address) && _.isUndefined(params.id)) {
            throw new Error('address or id of address required');
        }
        let query;
        if (params.address) {
            query = params.address;
        }
        else {
            query = params.id;
        }
        const url = this.url(`/address/${encodeURIComponent(query)}/tokenforward`);
        this._wallet = await this.bitgo.post(url).send(params).result();
        return this._wallet;
    }
    /**
     * Sweep funds for a wallet
     *
     * @param {Object} params - parameters object
     * @param {String} params.address - The address to send all the funds in the wallet to
     * @param {String} params.walletPassphrase - the users wallet passphrase
     * @param {String} params.xprv - the private key in string form if the walletPassphrase is not available
     * @param {String} params.otp - Two factor auth code to enable sending the transaction
     * @param {Number} params.feeTxConfirmTarget - Estimate the fees to aim for first confirmation within this number of blocks
     * @param {Number} params.feeRate - The desired fee rate for the transaction in satoshis/kB
     * @param {Number} [params.maxFeeRate] - upper limit for feeRate in satoshis/kB
     * @param {Boolean} [params.allowPartialSweep] - allows sweeping 200 unspents when the wallet has more than that
     * @returns txHex {String} the txHex of the signed transaction
     */
    async sweep(params = {}) {
        params = params || {};
        common.validateParams(params, ['address'], ['walletPassphrase', 'xprv', 'otp']);
        // The sweep API endpoint is only available to utxo-based coins
        if (!this.baseCoin.sweepWithSendMany()) {
            if (this.confirmedBalanceString() !== this.balanceString()) {
                throw new Error('cannot sweep when unconfirmed funds exist on the wallet, please wait until all inbound transactions confirm');
            }
            const value = await this.bitgo.get(this.url('/maximumSpendable')).result();
            const maximumSpendable = new bignumber_js_1.default(value.maximumSpendable);
            if (value === undefined || maximumSpendable.isZero()) {
                throw new Error('no funds to sweep');
            }
            const sendManyParams = {
                ...params,
                recipients: [
                    {
                        address: params.address || '', // Ensure address is always a string
                        amount: maximumSpendable.toString(),
                    },
                ],
            };
            return this.sendMany(sendManyParams);
        }
        // the following flow works for all UTXO coins
        const reqId = new utils_1.RequestTracer();
        const filteredParams = _.pick(params, [
            'address',
            'feeRate',
            'maxFeeRate',
            'feeTxConfirmTarget',
            'allowPartialSweep',
            'txFormat',
        ]);
        this.bitgo.setRequestTracer(reqId);
        const response = await this.bitgo.post(this.url('/sweepWallet')).send(filteredParams).result();
        const transaction = await this.baseCoin.explainTransaction(response);
        if (transaction?.outputs.length) {
            const invalidOutputAddress = transaction.outputs.find((output) => output.address !== params.address);
            if (invalidOutputAddress) {
                throw new Error(`invalid sweep destination ${invalidOutputAddress.address}, specified ${params.address}`);
            }
        }
        else {
            throw new Error('invalid transaction, no destination address');
        }
        const keychains = (await this.baseCoin.keychains().getKeysForSigning({ wallet: this, reqId }));
        const transactionParams = {
            ...params,
            txPrebuild: response,
            keychain: keychains[0],
            userKeychain: keychains[0],
            backupKeychain: keychains.length > 1 ? keychains[1] : null,
            bitgoKeychain: keychains.length > 2 ? keychains[2] : null,
            prv: params.xprv,
        };
        const signedTransaction = await this.signTransaction(transactionParams);
        const selectParams = _.pick(params, ['otp']);
        const finalTxParams = _.extend({}, signedTransaction, selectParams);
        this.bitgo.setRequestTracer(reqId);
        return this.sendTransaction(finalTxParams, reqId);
    }
    /**
     * Freeze a given wallet
     * @param params
     * @returns {*}
     */
    async freeze(params = {}) {
        common.validateParams(params, [], []);
        if (params.duration) {
            if (!_.isNumber(params.duration)) {
                throw new Error('invalid duration: should be number of seconds');
            }
        }
        return await this.bitgo.post(this.url('/freeze')).send(params).result();
    }
    /**
     * Update comment of a transfer
     * @param params
     * @returns {*}
     */
    async transferComment(params = {}) {
        common.validateParams(params, ['id'], ['comment']);
        return await this.bitgo
            .post(this.baseCoin.url('/wallet/' + this._wallet.id + '/transfer/' + params.id + '/comment'))
            .send(params)
            .result();
    }
    /**
     * List the addresses for a given wallet
     * @param params
     * @returns {*}
     */
    async addresses(params = {}) {
        common.validateParams(params, [], []);
        const query = {};
        if (params.mine) {
            query.mine = !!params.mine;
        }
        if (!_.isUndefined(params.prevId)) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            query.prevId = params.prevId;
        }
        if (params.sort) {
            if (!_.isNumber(params.sort)) {
                throw new Error('invalid sort argument, expecting number');
            }
            query.sort = params.sort;
        }
        if (params.limit) {
            if (!_.isNumber(params.limit)) {
                throw new Error('invalid limit argument, expecting number');
            }
            query.limit = params.limit;
        }
        if (params.labelContains) {
            if (!_.isString(params.labelContains)) {
                throw new Error('invalid labelContains argument, expecting string');
            }
            query.labelContains = params.labelContains;
        }
        if (!_.isUndefined(params.segwit)) {
            if (!_.isBoolean(params.segwit)) {
                throw new Error('invalid segwit argument, expecting boolean');
            }
            query.segwit = params.segwit;
        }
        if (!_.isUndefined(params.chains)) {
            if (!_.isArray(params.chains)) {
                throw new Error('invalid chains argument, expecting array of numbers');
            }
            query.chains = params.chains;
        }
        if (!_.isNil(params.includeBalances)) {
            if (!_.isBoolean(params.includeBalances)) {
                throw new Error('invalid includeBalances argument, expecting boolean');
            }
            query.includeBalances = params.includeBalances;
        }
        if (!_.isNil(params.includeTokens)) {
            if (!_.isBoolean(params.includeTokens)) {
                throw new Error('invalid includeTokens argument, expecting boolean');
            }
            query.includeTokens = params.includeTokens;
        }
        if (!_.isNil(params.includeTotalAddressCount)) {
            if (!_.isBoolean(params.includeTotalAddressCount)) {
                throw new Error('invalid includeTotalAddressCount argument, expecting boolean');
            }
            query.includeTotalAddressCount = params.includeTotalAddressCount;
        }
        if (params.returnBalancesForToken) {
            if (!_.isString(params.returnBalancesForToken)) {
                throw new Error('invalid returnBalancesForToken argument, expecting string');
            }
            query.returnBalancesForToken = params.returnBalancesForToken;
        }
        if (!_.isNil(params.pendingDeployment)) {
            if (!_.isBoolean(params.pendingDeployment)) {
                throw new Error('invalid pendingDeployment argument, expecting boolean');
            }
            query.pendingDeployment = params.pendingDeployment;
        }
        return this.bitgo
            .get(this.baseCoin.url('/wallet/' + this._wallet.id + '/addresses'))
            .query(query)
            .result();
    }
    /**
     * List the addresses sorted by balance for a given wallet
     * @param params
     * @returns {*}
     */
    async addressesByBalance(params) {
        const query = {
            token: params.token,
            nftCollectionId: params.nftCollectionId,
            nftId: params.nftId,
        };
        query.sort = params.sort ?? -1;
        query.page = params.page ?? 1;
        query.limit = params.limit ?? 500;
        if (!_.isNumber(query.sort)) {
            throw new Error('invalid sort argument, expecting number');
        }
        if (!_.isNumber(query.page)) {
            throw new Error('invalid page argument, expecting number');
        }
        if (!_.isNumber(query.limit)) {
            throw new Error('invalid limit argument, expecting number');
        }
        if (query.limit < 1 || query.limit > 500) {
            throw new Error('limit argument must be between 1 and 500');
        }
        // Assert that either both nftCollectionId and nftId are provided or neither are provided
        const hasNftCollectionId = !_.isUndefined(query.nftCollectionId);
        const hasNftId = !_.isUndefined(query.nftId);
        if (hasNftCollectionId !== hasNftId) {
            throw new Error('nftCollectionId and nftId must both be provided or both be omitted');
        }
        return this.bitgo
            .get(this.baseCoin.url('/wallet/' + this._wallet.id + '/addresses/balances'))
            .query(query)
            .result();
    }
    /**
     * Get a single wallet address by its id
     * @param params
     * @returns {*}
     */
    async getAddress(params = {}) {
        common.validateParams(params, [], ['address', 'id']);
        let query;
        if (_.isUndefined(params.address) && _.isUndefined(params.id)) {
            throw new Error('address or id of address required');
        }
        if (params.address) {
            query = params.address;
        }
        else {
            query = params.id;
        }
        if (params.reqId) {
            this.bitgo.setRequestTracer(params.reqId);
        }
        return this.bitgo
            .get(this.baseCoin.url(`/wallet/${this._wallet.id}/address/${encodeURIComponent(query)}`))
            .result();
    }
    /**
     * Create one or more new address(es) for use with this wallet.
     *
     * If the `count` field is defined and greater than 1, an object with a single
     * array property named `addresses` containing `count` address objects
     * will be returned. Otherwise, a single address object is returned.
     *
     * @param params
     * @param {Number} params.chain on which the new address should be created
     * @param {(Number|String)} params.gasPrice gas price for new address creation, if applicable
     * @param {String} params.label label for the new address(es)
     * @param {Number} params.count=1 number of new addresses which should be created (maximum 250)
     * @param {Number} params.forwarderVersion The version of address to create, if applicable
     * @param {Boolean} params.lowPriority Ethereum-specific param to create address using low priority fee address
     * @param {String} params.baseAddress base address of the wallet(optional parameter)
     * @param {Boolean} params.allowSkipVerifyAddress When set to false, it throws error if address verification is skipped for any reason. Default is true.
     * @param {String} params.onToken mandatory in case of the OFC wallet, the name of token to create address for
     * Address verification can be skipped when forwarderVersion is 0 and pendingChainInitialization is true OR
     * if 'coinSpecific' is not part of the response from api call to create address
     */
    async createAddress(params = {}) {
        const addressParams = {};
        const reqId = new utils_1.RequestTracer();
        const { chain, gasPrice, label, lowPriority, forwarderVersion, format, count = 1, baseAddress, allowSkipVerifyAddress = true, onToken, evmKeyRingReferenceAddress, } = params;
        if (!_.isUndefined(chain)) {
            if (!_.isInteger(chain)) {
                throw new Error('chain has to be an integer');
            }
            addressParams.chain = chain;
        }
        if (!_.isUndefined(gasPrice)) {
            if (!_.isInteger(gasPrice) && (isNaN(Number(gasPrice)) || !_.isString(gasPrice))) {
                throw new Error('gasPrice has to be an integer or numeric string');
            }
            addressParams.gasPrice = gasPrice;
        }
        if (!_.isUndefined(forwarderVersion)) {
            if (!_.isInteger(forwarderVersion) || forwarderVersion < 0 || forwarderVersion > 5) {
                throw new Error('forwarderVersion has to be an integer 0, 1, 2, 3, 4 or 5');
            }
            addressParams.forwarderVersion = forwarderVersion;
        }
        if (!_.isUndefined(label)) {
            if (!_.isString(label)) {
                throw new Error('label has to be a string');
            }
            addressParams.label = label;
        }
        if (!_.isUndefined(baseAddress)) {
            if (!_.isString(baseAddress)) {
                throw new Error('baseAddress has to be a string');
            }
        }
        if (!_.isUndefined(allowSkipVerifyAddress)) {
            if (!_.isBoolean(allowSkipVerifyAddress)) {
                throw new Error('allowSkipVerifyAddress has to be a boolean');
            }
        }
        if (!_.isInteger(count) || count <= 0 || count > 250) {
            throw new Error('count has to be a number between 1 and 250');
        }
        if (!_.isUndefined(lowPriority)) {
            if (!_.isBoolean(lowPriority)) {
                throw new Error('lowPriority has to be a boolean');
            }
            addressParams.lowPriority = lowPriority;
        }
        if (!_.isUndefined(format)) {
            if (!_.isString(format)) {
                throw new Error('format has to be a string');
            }
            addressParams.format = format;
        }
        if (this.baseCoin.getFamily() === 'ofc') {
            if (!_.isUndefined(onToken)) {
                if (!_.isString(onToken)) {
                    throw new Error('onToken has to be a string');
                }
                addressParams.onToken = onToken;
            }
            else {
                throw new Error('onToken is a mandatory parameter for OFC wallets');
            }
            if (!_.isString(onToken)) {
                throw new Error('onToken has to be a string');
            }
        }
        if (!_.isUndefined(evmKeyRingReferenceAddress)) {
            if (!_.isString(evmKeyRingReferenceAddress)) {
                throw new Error('evmKeyRingReferenceAddress has to be a string');
            }
            if (!this.baseCoin.isEVM()) {
                throw new Error('evmKeyRingReferenceAddress is only supported for EVM chains');
            }
            if (!this.baseCoin.isValidAddress(evmKeyRingReferenceAddress)) {
                throw new Error('evmKeyRingReferenceAddress must be a valid address');
            }
            addressParams.evmKeyRingReferenceAddress = evmKeyRingReferenceAddress;
        }
        // get keychains for address verification
        const keychains = await Promise.all(this._wallet.keys.map((k) => this.baseCoin.keychains().get({ id: k, reqId })));
        const rootAddress = _.get(this._wallet, 'receiveAddress.address');
        const newAddresses = _.times(count, async () => {
            this.bitgo.setRequestTracer(reqId);
            const newAddress = (await this.bitgo
                .post(this.baseCoin.url('/wallet/' + this._wallet.id + '/address'))
                .send(addressParams)
                .result());
            newAddress.keychains = keychains;
            newAddress.baseAddress = baseAddress ?? _.get(this._wallet, 'coinSpecific.baseAddress');
            newAddress.format = addressParams.format;
            const verificationData = _.merge({}, newAddress, {
                rootAddress,
                walletVersion: _.get(this._wallet, 'coinSpecific.walletVersion'),
            });
            if (verificationData.error) {
                throw new errors_1.AddressGenerationError(verificationData.error);
            }
            verificationData.impliedForwarderVersion = forwarderVersion ?? verificationData.coinSpecific?.forwarderVersion;
            // For SMC (Self-Managed Custodial) wallets, pass derivedFromParentWithSeed from user keychain
            // The verification function will compute the derivation prefix internally
            // Custodial wallets don't need this as their commonKeychain already accounts for the prefix
            if (this.multisigType() === 'tss' && this.type() === 'cold' && this._wallet.keys.length > keychain_1.KeyIndices.USER) {
                const userKeychain = keychains[keychain_1.KeyIndices.USER];
                verificationData.derivedFromParentWithSeed = userKeychain?.derivedFromParentWithSeed;
            }
            // This condition was added in first place because in celo, when verifyAddress method was called on addresses which were having pendingChainInitialization as true, it used to throw some error
            // In case of forwarder version 1 eth addresses, addresses need to be verified even if the pendingChainInitialization flag is true
            if (verificationData.coinSpecific &&
                (!verificationData.coinSpecific.pendingChainInitialization || verificationData.impliedForwarderVersion === 1)) {
                // can't verify addresses which are pending chain initialization, as the address is hidden
                let isWalletAddress = false;
                try {
                    isWalletAddress = await this.baseCoin.isWalletAddress(verificationData, this);
                }
                catch (e) {
                    if (!(e instanceof errors_1.MethodNotImplementedError)) {
                        throw e;
                    }
                    // FIXME(BG-43225): implement this correctly
                    isWalletAddress = true;
                }
                if (!isWalletAddress) {
                    throw new Error(`not a wallet address`);
                }
            }
            else if (!allowSkipVerifyAddress) {
                throw new Error(`address verification skipped for count = ${count}`);
            }
            return newAddress;
        });
        if (newAddresses.length === 1) {
            return newAddresses[0];
        }
        return {
            addresses: await Promise.all(newAddresses),
        };
    }
    /**
     * Update properties on an address
     * @param params
     * @returns {*}
     */
    async updateAddress(params = {}) {
        const address = params.address;
        if (!_.isString(address)) {
            throw new Error('missing required string parameter address');
        }
        const putParams = _.pick(params, ['label']);
        const url = this.url('/address/' + encodeURIComponent(address));
        return this.bitgo.put(url).send(putParams).result();
    }
    /**
     * Get account resources for the given addresses
     * @param params - options object containing addresses and optional destinationAddress
     * @returns {Promise<GetAccountResourcesResponse>} - response from WP API
     */
    async getAccountResources(params) {
        const { addresses, destinationAddress } = params;
        if (!Array.isArray(addresses)) {
            throw new Error('addresses must be an array');
        }
        if (addresses.length === 0) {
            throw new Error('addresses array cannot be empty');
        }
        const query = { addresses };
        if (destinationAddress) {
            query.destinationAddress = destinationAddress;
        }
        return this.bitgo.get(this.url('/getAccountResources')).query(query).result();
    }
    async updateWalletBuildDefaults(params) {
        common.validateParams(params, [], ['minFeeRate', 'changeAddressType', 'txFormat']);
        return this.updateWallet({
            buildDefaults: {
                minFeeRate: params.minFeeRate,
                changeAddressType: params.changeAddressType,
                txFormat: params.txFormat,
            },
        });
    }
    async updateWallet(params) {
        const filteredParams = _.pick(params, ['coinSpecific', 'buildDefaults', 'label']);
        const { coinSpecific, ...rest } = filteredParams;
        return this.bitgo
            .put(this.url())
            .send({
            ...rest,
            ...(coinSpecific
                ? {
                    coinSpecific: {
                        [this.baseCoin.getChain()]: coinSpecific,
                    },
                }
                : {}),
        })
            .result();
    }
    /**
     * List webhooks on this wallet
     * @param params
     */
    async listWebhooks(params = {}) {
        const query = {};
        if (params.prevId) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            query.prevId = params.prevId;
        }
        if (params.limit) {
            if (!_.isNumber(params.limit)) {
                throw new Error('invalid limit argument, expecting number');
            }
            query.limit = params.limit;
        }
        return this.bitgo.get(this.url('/webhooks')).query(query).result();
    }
    /**
     * Simulate wallet webhook, currently for webhooks of type transfer and pending approval
     * @param params
     * - webhookId (required) id of the webhook to be simulated
     * - transferId (optional but required for transfer webhooks) id of the simulated transfer
     * - pendingApprovalId (optional but required for pending approval webhooks) id of the simulated pending approval
     * @returns {*}
     */
    async simulateWebhook(params = {}) {
        common.validateParams(params, ['webhookId'], ['transferId', 'pendingApprovalId']);
        const hasTransferId = !!params.transferId;
        const hasPendingApprovalId = !!params.pendingApprovalId;
        if (!hasTransferId && !hasPendingApprovalId) {
            throw new Error('must supply either transferId or pendingApprovalId');
        }
        if (hasTransferId && hasPendingApprovalId) {
            throw new Error('must supply either transferId or pendingApprovalId, but not both');
        }
        // depending on the coin type of the wallet, the txHash has to adhere to its respective format
        // but the server takes care of that
        // only take the transferId and pendingApprovalId properties
        const filteredParams = _.pick(params, ['transferId', 'pendingApprovalId']);
        const webhookId = params.webhookId;
        return this.bitgo
            .post(this.url('/webhooks/' + webhookId + '/simulate'))
            .send(filteredParams)
            .result();
    }
    /**
     * Add a webhook to this wallet
     * @param params
     */
    async addWebhook(params = {}) {
        common.validateParams(params, ['url', 'type'], []);
        return this.bitgo.post(this.url('/webhooks')).send(params).result();
    }
    /**
     * Remove a webhook from this wallet
     * @param params
     */
    async removeWebhook(params = {}) {
        common.validateParams(params, ['url', 'type'], []);
        return this.bitgo.del(this.url('/webhooks')).send(params).result();
    }
    /**
     * Gets the user keychain for this wallet
     *
     * The user keychain is the first keychain of the wallet and usually has the encrypted prv stored on BitGo.
     * Useful when trying to get the users' keychain from the server before decrypting to sign a transaction.
     */
    async getEncryptedUserKeychain() {
        const tryKeyChain = async (index) => {
            if (!this._wallet.keys || index >= this._wallet.keys.length) {
                throw new errors_1.MissingEncryptedKeychainError();
            }
            const params = { id: this._wallet.keys[index] };
            const keychain = await this.baseCoin.keychains().get(params);
            // If we find the prv, then this is probably the user keychain we're looking for
            if (keychain.encryptedPrv) {
                return keychain;
            }
            return tryKeyChain(index + 1);
        };
        return tryKeyChain(0);
    }
    /**
     * Gets the unencrypted private key for this wallet (be careful!)
     * Requires wallet passphrase
     *
     * @param params
     */
    async getPrv(params = {}) {
        common.validateParams(params, [], ['walletPassphrase', 'prv']);
        // Prepare signing key
        if (_.isUndefined(params.prv) && _.isUndefined(params.walletPassphrase)) {
            throw new Error('must either provide prv or wallet passphrase');
        }
        if (!_.isUndefined(params.prv) && !_.isString(params.prv)) {
            throw new Error('prv must be a string');
        }
        if (!_.isUndefined(params.walletPassphrase) && !_.isString(params.walletPassphrase)) {
            throw new Error('walletPassphrase must be a string');
        }
        if (params.prv) {
            return params.prv;
        }
        const userKeychain = await this.getEncryptedUserKeychain();
        if (!params.walletPassphrase) {
            throw new Error('wallet passphrase was not provided');
        }
        const userPrv = (0, keychain_1.decryptKeychainPrivateKey)(this.bitgo, userKeychain, params.walletPassphrase);
        if (!userPrv) {
            throw new Error('error decrypting wallet private key');
        }
        return userPrv;
    }
    isMultiUserKeyWallet() {
        return this._wallet.coinSpecific?.features?.includes('multi-key-user-key') ?? false;
    }
    /**
     * Send an encrypted wallet share to BitGo.
     * @param params
     */
    async createShare(params = {}) {
        common.validateParams(params, ['user', 'permissions'], []);
        const isMultiUserKeyWallet = this.isMultiUserKeyWallet();
        if (isMultiUserKeyWallet) {
            if (params.keychain && !_.isEmpty(params.keychain)) {
                throw new Error('keychain property must not be provided for multi-key-user-key wallets');
            }
            // Remove keychain from params if presents
            return await this.bitgo.post(this.url('/share')).send(_.omit(params, 'keychain')).result();
        }
        if (params.keychain && !_.isEmpty(params.keychain)) {
            if (!params.keychain.pub ||
                !params.keychain.encryptedPrv ||
                !params.keychain.fromPubKey ||
                !params.keychain.toPubKey ||
                !params.keychain.path) {
                throw new Error('requires keychain parameters - pub, encryptedPrv, fromPubKey, toPubKey, path');
            }
        }
        return this.bitgo.post(this.url('/share')).send(params).result();
    }
    /**
     * Shares a wallet with multiple users by creating bulk wallet shares.
     *
     * @async
     * @param {BulkWalletShareOptions} params - The options for sharing wallets in bulk.
     * @param {Array<ShareOption>} params.shareOptions - An array of share option objects containing user and permissions information.
     * @param {Object} [params.shareOptions[].keychain] - The keychain object used to share the wallet.
     * @param {string} [params.shareOptions[].keychain.toPubKey] - The recipient's public key.
     * @param {string} [params.shareOptions[].keychain.path] - The derivation path of the keychain.
     * @param {string} params.shareOptions[].user - The user to share the wallet with.
     * @param {string} params.shareOptions[].permissions - The permissions granted to the user.
     * @param {string} [params.walletPassphrase] - The wallet passphrase used to decrypt the keychain.
     * @throws {Error} If `shareOptions` is empty, or if required keychain parameters (`toPubKey` and `path`) are missing when needed.
     * @throws {Error} If unable to decrypt the user keychain.
     * @returns {Promise<CreateBulkWalletShareListResponse>} A promise that resolves with the response of the bulk wallet share creation.
     */
    async createBulkWalletShare(params) {
        if (params.keyShareOptions.length === 0) {
            throw new Error('No share options provided');
        }
        const bulkCreateShareOptions = [];
        // Check if any share option needs a keychain (has 'spend' permission)
        const anyNeedsKeychain = params.keyShareOptions.some((opt) => opt.permissions && opt.permissions.includes('spend'));
        // Fetch and decrypt the keychain ONCE for all users
        let decryptedKeychain;
        if (anyNeedsKeychain) {
            try {
                decryptedKeychain = await this.getDecryptedKeychainForSharing(params.walletPassphrase);
            }
            catch (e) {
                if (e instanceof errors_1.MissingEncryptedKeychainError) {
                    decryptedKeychain = undefined;
                }
                else {
                    throw e;
                }
            }
        }
        for (const shareOption of params.keyShareOptions) {
            try {
                common.validateParams(shareOption, ['userId', 'pubKey', 'path'], []);
            }
            catch (e) {
                if (!shareOption.pubKey) {
                    throw new errors_1.NeedUserSignupError(shareOption.userId);
                }
                throw e;
            }
            const needsKeychain = shareOption.permissions && shareOption.permissions.includes('spend');
            if (needsKeychain && decryptedKeychain) {
                const sharedKeychain = this.encryptPrvForUser(decryptedKeychain.prv, decryptedKeychain.pub, shareOption.pubKey, shareOption.path);
                bulkCreateShareOptions.push({
                    user: shareOption.userId,
                    permissions: shareOption.permissions,
                    keychain: sharedKeychain,
                });
            }
        }
        return await this.createBulkKeyShares(bulkCreateShareOptions);
    }
    /**
     * Creates bulk wallet share entries for specified share options.
     * Filters out share options that do not contain valid keychain information or have missing keychain fields.
     * If all share options are invalid or empty, it throws an error.
     * Sends a POST request to create the wallet shares for valid share options.
     *
     * @async
     * @param {BulkCreateShareOption[]} [params=[]] - The array of share options to process.
     *   Keychain entries must include the following fields: `pub`, `encryptedPrv`, `fromPubKey`, `toPubKey`, and `path`.
     * @returns {Promise<CreateBulkWalletShareListResponse>} A promise resolving to the result of the wallet shares creation.
     * @throws {Error} Throws an error if no valid share options are provided.
     */
    async createBulkKeyShares(params = []) {
        params = params.filter((shareOption) => {
            try {
                common.validateParams(shareOption.keychain, ['pub', 'encryptedPrv', 'fromPubKey', 'toPubKey', 'path'], []);
                return true;
            }
            catch (e) {
                // Exclude share options with invalid keychain
                return false;
            }
        });
        if (!params || Object.keys(params).length === 0) {
            throw new Error('shareOptions cannot be empty');
        }
        const url = this.bitgo.url(`/wallet/${this._wallet.id}/walletshares`, 2);
        return this.bitgo.post(url).send({ shareOptions: params }).result();
    }
    /**
     * Gets keychain with encrypted private key to be shared for wallet sharing.
     */
    async getEncryptedWalletKeychainForWalletSharing() {
        if (this.baseCoin.getFamily() === 'lnbtc') {
            // lightning coin does not use user key to sign the transactions from SDK.
            // it uses user auth key instead.
            return await (0, lightningWalletUtil_1.getLightningAuthKey)(this, 'userAuth');
        }
        else {
            return await this.getEncryptedUserKeychain();
        }
    }
    /**
     * Fetches and decrypts the wallet keychain for sharing.
     * This method fetches the keychain once and decrypts it, returning the decrypted
     * private key and public key info needed for sharing with multiple users.
     *
     * @param walletPassphrase - The passphrase to decrypt the keychain
     * @returns Object containing decrypted prv and pub, or undefined for cold wallets
     */
    async getDecryptedKeychainForSharing(walletPassphrase) {
        const keychain = await this.getEncryptedWalletKeychainForWalletSharing();
        if (!keychain.encryptedPrv) {
            return undefined;
        }
        if (!walletPassphrase) {
            throw new Error('Missing walletPassphrase argument');
        }
        const prv = (0, keychain_1.decryptKeychainPrivateKey)(this.bitgo, keychain, walletPassphrase);
        if (!prv) {
            throw new errors_1.IncorrectPasswordError('Password shared is incorrect for this wallet');
        }
        // Only one of pub/commonPub/commonKeychain should be present in the keychain
        let pub = keychain.pub ?? keychain.commonPub;
        if (keychain.commonKeychain) {
            pub =
                this.baseCoin.getMPCAlgorithm() === 'eddsa'
                    ? eddsa_1.default.getPublicKeyFromCommonKeychain(keychain.commonKeychain)
                    : ecdsa_1.EcdsaUtils.getPublicKeyFromCommonKeychain(keychain.commonKeychain);
        }
        if (!pub) {
            throw new Error('Unable to determine public key from keychain');
        }
        return { prv, pub };
    }
    /**
     * Encrypts a decrypted private key for sharing with a specific user.
     * This is the pure encryption step - no API calls, no decryption.
     *
     * @param decryptedPrv - The already-decrypted private key
     * @param pub - The wallet's public key
     * @param userPubkey - The recipient user's public key
     * @param path - The key path
     * @returns The encrypted keychain for the recipient with all required fields
     */
    encryptPrvForUser(decryptedPrv, pub, userPubkey, path) {
        const eckey = (0, bitcoin_1.makeRandomKey)();
        const secret = (0, ecdh_1.getSharedSecret)(eckey, Buffer.from(userPubkey, 'hex')).toString('hex');
        const newEncryptedPrv = this.bitgo.encrypt({ password: secret, input: decryptedPrv });
        const keychain = {
            pub,
            encryptedPrv: newEncryptedPrv,
            fromPubKey: eckey.publicKey.toString('hex'),
            toPubKey: userPubkey,
            path: path,
        };
        (0, assert_1.default)(keychain.pub, 'pub must be defined for sharing');
        (0, assert_1.default)(keychain.encryptedPrv, 'encryptedPrv must be defined for sharing');
        (0, assert_1.default)(keychain.fromPubKey, 'fromPubKey must be defined for sharing');
        (0, assert_1.default)(keychain.toPubKey, 'toPubKey must be defined for sharing');
        (0, assert_1.default)(keychain.path, 'path must be defined for sharing');
        return keychain;
    }
    /**
     * Prepares a keychain for sharing with another user.
     * Fetches the wallet keychain, decrypts it, and encrypts it for the recipient.
     *
     * @param walletPassphrase - The passphrase to decrypt the keychain
     * @param pubkey - The recipient's public key
     * @param path - The key path
     * @returns The encrypted keychain for the recipient
     */
    async prepareSharedKeychain(walletPassphrase, pubkey, path) {
        try {
            const decryptedKeychain = await this.getDecryptedKeychainForSharing(walletPassphrase);
            if (!decryptedKeychain) {
                return {};
            }
            return this.encryptPrvForUser(decryptedKeychain.prv, decryptedKeychain.pub, pubkey, path);
        }
        catch (e) {
            if (e instanceof errors_1.MissingEncryptedKeychainError) {
                // ignore this error because this looks like a cold wallet
                return {};
            }
            throw e;
        }
    }
    /**
     * Share this wallet with another BitGo user.
     * @param params
     * @returns {*}
     */
    async shareWallet(params = {}) {
        common.validateParams(params, ['email', 'permissions'], ['walletPassphrase', 'message']);
        if (params.reshare !== undefined && !_.isBoolean(params.reshare)) {
            throw new Error('Expected reshare to be a boolean.');
        }
        if (params.skipKeychain !== undefined && !_.isBoolean(params.skipKeychain)) {
            throw new Error('Expected skipKeychain to be a boolean. ');
        }
        // Check if this is a multi-key-user-key OFC wallet
        // For multi-key-user-key wallets, skip keychain preparation regardless of other conditions
        const needsKeychain = !this.isMultiUserKeyWallet() &&
            !params.skipKeychain &&
            params.permissions &&
            params.permissions.indexOf('spend') !== -1;
        if (params.disableEmail !== undefined && !_.isBoolean(params.disableEmail)) {
            throw new Error('Expected disableEmail to be a boolean.');
        }
        if (!_.isString(params.email)) {
            throw new Error('missing required string parameter email');
        }
        const sharing = (await this.bitgo.getSharingKey({
            email: params.email.toLowerCase(),
        }));
        let sharedKeychain;
        if (needsKeychain) {
            sharedKeychain = await this.prepareSharedKeychain(params.walletPassphrase, sharing.pubkey, sharing.path);
        }
        const options = {
            user: sharing.userId,
            permissions: params.permissions,
            reshare: params.reshare,
            message: params.message,
            disableEmail: params.disableEmail,
            skipKeychain: Object.keys(sharedKeychain ?? {}).length === 0,
            keychain: Object.keys(sharedKeychain ?? {}).length === 0 ? undefined : sharedKeychain,
        };
        return await this.createShare(options);
    }
    /**
     * Remove user from wallet
     * @param params
     * - userId Id of the user to remove
     * @return {*}
     */
    async removeUser(params = {}) {
        common.validateParams(params, ['userId'], []);
        const userId = params.userId;
        return await this.bitgo.del(this.url('/user/' + userId)).result();
    }
    /**
     * Fetch a transaction prebuild (unsigned transaction) from BitGo
     *
     * @param {Object} params
     * @param {{address: string, amount: string}} params.recipients - list of recipients and necessary recipient information
     * @param {Number} params.numBlocks - Estimates the approximate fee per kilobyte necessary for a transaction confirmation within numBlocks blocks
     * @param {Number} params.feeRate - the desired feeRate for the transaction in base units/kB
     * @param {Number} params.maxFeeRate - upper limit for feeRate in base units/kB
     * @param {Number} params.minConfirms - Minimum number of confirmations unspents going into this transaction should have
     * @param {Boolean} params.enforceMinConfirmsForChange - Enforce minimum number of confirmations on change (internal) inputs.
     * @param {Number} params.targetWalletUnspents - The desired count of unspents in the wallet. If the wallet’s current unspent count is lower than the target, up to four additional change outputs will be added to the transaction.
     * @param {Number | String} params.minValue - Ignore unspents smaller than this amount of base units
     * @param {Number | String} params.maxValue - Ignore unspents larger than this amount of base units
     * @param {Number} params.sequenceId - The sequence ID of the transaction
     * @param {Number} params.lastLedgerSequence - Absolute max ledger the transaction should be accepted in, whereafter it will be rejected.
     * @param {Number} params.ledgerSequenceDelta - Relative ledger height (in relation to the current ledger) that the transaction should be accepted in, whereafter it will be rejected.
     * @param {Number} params.gasPrice - Custom gas price to be used for sending the transaction
     * @param {Number} params.gasLimit - Custom gas limit to be used for sending the transaction
     * @param {Boolean} params.noSplitChange - Set to true to disable automatic change splitting for purposes of unspent management
     * @param {Array} params.unspents - The unspents to use in the transaction. Each unspent should be in the form prevTxId:nOutput
     * @param {String} params.changeAddress - Specifies the destination of the change output
     * @param {Boolean} params.nonParticipation - (Algorand) Non participating key reg transaction
     * @param {Number} params.validFromBlock - (Algorand) The minimum round this will run on
     * @param {Number} params.validToBlock - (Algorand) The maximum round this will run on
     * @param {Boolean} params.instant - Build this transaction to conform with instant sending coin-specific method (if available)
     * @param {Boolean} params.keepAlive - (Polkadot) keep address alive by sending the address minimum funding amount, used during wallet consolidation, true by default
     * @param {{value: String, type: String}} params.memo - Memo to use in transaction (supported by Stellar)
     * @param {String} param.transferId - transfer Id to use in transaction (supported by casper)
     * @param {String} params.addressType - The type of address to create for change. One of `p2sh`, `p2shP2wsh`, and `p2wsh`. Case-sensitive.
     * @param {Boolean} params.hop - Build this as an Ethereum hop transaction
     * @param {Object} params.reservation - Object to reserve the unspents that this tx build uses. Format is reservation = { expireTime: ISODateString, pendingApprovalId: String }
     * @param {String} params.walletPassphrase The passphrase to the wallet user key, to sign commitment data for Ethereum hop transactions
     * @param {String} params.walletContractAddress - The contract address used as the "to" field of a transaction
     * @returns {*}
     */
    async prebuildTransaction(params = {}) {
        if (this._wallet.multisigType === 'tss') {
            return this.prebuildTransactionTxRequests(params);
        }
        // Whitelist params to build tx
        const whitelistedParams = this.baseCoin.preprocessBuildParams(_.pick(params, this.prebuildWhitelistedParams()));
        debug('prebuilding transaction: %O', whitelistedParams);
        if (params.reqId) {
            this.bitgo.setRequestTracer(params.reqId);
        }
        const extraParams = await this.baseCoin.getExtraPrebuildParams(Object.assign(params, { wallet: this }));
        Object.assign(whitelistedParams, extraParams);
        const queryParams = {
            offlineVerification: params.offlineVerification ? true : undefined,
        };
        const buildQuery = this.bitgo
            .post(this.baseCoin.url('/wallet/' + this.id() + '/tx/build'))
            .query(queryParams)
            .send(whitelistedParams)
            .result();
        const blockHeightQuery = _.isFunction(this.baseCoin.getLatestBlockHeight)
            ? this.baseCoin.getLatestBlockHeight(params.reqId)
            : Promise.resolve(undefined);
        const queries = [buildQuery, blockHeightQuery];
        const [buildResponse, blockHeight] = (await Promise.all(queries));
        debug('postprocessing transaction prebuild: %O', buildResponse);
        if (!_.isUndefined(blockHeight)) {
            buildResponse.blockHeight = blockHeight;
        }
        let prebuild = (await this.baseCoin.postProcessPrebuild(Object.assign(buildResponse, {
            wallet: this,
            buildParams: whitelistedParams,
        })));
        delete prebuild.wallet;
        delete prebuild.buildParams;
        prebuild = _.extend({}, prebuild, { walletId: this.id() });
        if (this._wallet && this._wallet.coinSpecific && !params.walletContractAddress) {
            prebuild = _.extend({}, prebuild, {
                walletContractAddress: this._wallet.coinSpecific.baseAddress,
            });
        }
        prebuild = _.extend({}, prebuild, { reqId: params.reqId });
        debug('final transaction prebuild: %O', prebuild);
        return prebuild;
    }
    /**
     * Gets the User Keychain and sign a TSS transaction
     * @param txRequestId The transaction request id
     * @param walletPassphrase The wallet passphrase
     * @return Promise<SignedTransaction>
     */
    async getUserKeyAndSignTssTransaction({ txRequestId, walletPassphrase, }) {
        if (this._wallet.multisigType !== 'tss') {
            throw new Error('getUserKeyAndSignTssTransaction is only supported for TSS wallets');
        }
        const reqId = new utils_1.RequestTracer();
        // Doing a sanity check for password here to avoid doing further work if we know it's wrong
        const keychains = await this.getKeychainsAndValidatePassphrase({
            reqId,
            walletPassphrase,
        });
        const userKeychain = keychains[0];
        if (!userKeychain || !userKeychain.encryptedPrv) {
            throw new Error('the user keychain does not have property encryptedPrv');
        }
        return this.signTransaction({
            txPrebuild: { txRequestId },
            walletPassphrase,
            reqId,
            keychain: userKeychain,
        });
    }
    /**
     * Sign a transaction
     * @param params
     * - txPrebuild
     * - [keychain / key] (object) or prv (string)
     * - walletPassphrase
     * - verifyTxParams (optional) - when provided, the transaction will be verified before signing
     *   - txParams: transaction parameters used for verification
     *   - verification: optional verification options
     * @return {*}
     */
    async signTransaction(params = {}) {
        const { txPrebuild, apiVersion, txRequestId } = params;
        if (_.isFunction(params.customCommitmentGeneratingFunction) &&
            _.isFunction(params.customGShareGeneratingFunction) &&
            _.isFunction(params.customRShareGeneratingFunction)) {
            // invoke external signer TSS for EdDSA workflow
            return this.signTransactionTssExternalSignerEdDSA(params, this.baseCoin);
        }
        if (_.isFunction(params.customPaillierModulusGeneratingFunction) &&
            _.isFunction(params.customKShareGeneratingFunction) &&
            _.isFunction(params.customMuDeltaShareGeneratingFunction) &&
            _.isFunction(params.customSShareGeneratingFunction)) {
            // invoke external signer TSS for ECDSA workflow
            return this.signTransactionTssExternalSignerECDSA(this.baseCoin, params);
        }
        if (_.isFunction(params.customMPCv2SigningRound1GenerationFunction) &&
            _.isFunction(params.customMPCv2SigningRound2GenerationFunction) &&
            _.isFunction(params.customMPCv2SigningRound3GenerationFunction)) {
            // invoke external signer TSS for ECDSA MPCv2workflow
            return this.signTransactionTssExternalSignerECDSAMPCv2(this.baseCoin, params);
        }
        if (!txPrebuild || typeof txPrebuild !== 'object') {
            if (this.multisigType() === 'onchain') {
                throw new Error('txPrebuild is required for on-chain multisig wallets');
            }
            if (!txRequestId) {
                throw new Error('txPrebuild or txRequestId is required for TSS wallets');
            }
            // We only do this if we're not using the external signer TSS flow
            params.txPrebuild = { txRequestId };
        }
        // Verify transaction if verifyTxParams is provided
        if (params.verifyTxParams && txPrebuild?.txHex) {
            const verifyParams = {
                txPrebuild: { ...txPrebuild },
                txParams: params.verifyTxParams.txParams,
                wallet: this,
                verification: params.verifyTxParams.verification,
                reqId: params.reqId,
                walletType: this.multisigType(),
            };
            await this.baseCoin.verifyTransaction(verifyParams);
        }
        if (params.walletPassphrase &&
            !(params.keychain || params.key) &&
            (this.type() === 'hot' || this.type() === undefined)) {
            // this logic should only apply to hot wallets
            if (!_.isString(params.walletPassphrase)) {
                throw new Error('walletPassphrase must be a string');
            }
            const keychains = await this.getKeychainsAndValidatePassphrase({
                reqId: params.reqId,
                walletPassphrase: params.walletPassphrase,
            });
            const userKeychain = keychains[0];
            if (!userKeychain || !userKeychain.encryptedPrv) {
                throw new Error('the user keychain does not have property encryptedPrv');
            }
            params.keychain = userKeychain;
        }
        const presign = await this.baseCoin.presignTransaction({
            ...params,
            walletData: this._wallet,
            tssUtils: this.tssUtils,
        });
        if (this.multisigType() === 'tss') {
            return this.signTransactionTss({
                ...presign,
                prv: this.getUserPrv(presign),
                apiVersion,
            });
        }
        let { pubs } = params;
        if (!pubs && this.baseCoin.keyIdsForSigning().length > 1) {
            const keychains = await this.baseCoin.keychains().getKeysForSigning({ wallet: this });
            pubs = keychains.map((k) => {
                (0, assert_1.default)(k.pub);
                return k.pub;
            });
        }
        const signTransactionParams = {
            ...presign,
            txPrebuild: { ...txPrebuild, walletId: this.id() },
            pubs,
            coin: this.baseCoin,
        };
        if (_.isFunction(params.customSigningFunction)) {
            if (typeof this.baseCoin.signWithCustomSigningFunction === 'function') {
                (0, assert_1.default)(pubs, 'pubs are required for custom signing');
                return this.baseCoin.signWithCustomSigningFunction(params.customSigningFunction, {
                    ...signTransactionParams,
                    pubs,
                });
            }
            const keys = await this.baseCoin.keychains().getKeysForSigning({ wallet: this });
            const signTransactionParamsWithSeed = {
                ...signTransactionParams,
                derivationSeed: keys[0]?.derivedFromParentWithSeed,
            };
            return params.customSigningFunction(signTransactionParamsWithSeed);
        }
        return this.baseCoin.signTransaction({
            ...signTransactionParams,
            prv: this.getUserPrv(presign),
            wallet: this,
        });
    }
    /**
     * Sign a typed structured data using TSS
     * @param params
     */
    async signTypedData(params) {
        if (!this.baseCoin.supportsSigningTypedData()) {
            throw new Error(`Sign typed data not supported for ${this.baseCoin.getFullName()}`);
        }
        if (!params.typedData) {
            throw new Error(`Typed data required`);
        }
        if (this._wallet.multisigType !== 'tss') {
            throw new Error('Message signing only supported for TSS wallets');
        }
        if (_.isFunction(params.typedData.typedDataRaw)) {
            throw new Error('typedData.typedDataRaw must be JSON string');
        }
        if (_.isFunction(this.baseCoin.encodeTypedData)) {
            params.typedData.typedDataEncoded = this.baseCoin.encodeTypedData(params.typedData);
        }
        const keychains = await this.baseCoin.keychains().getKeysForSigning({ wallet: this, reqId: params.reqId });
        const userPrvOptions = {
            ...params,
            keychain: keychains[0],
        };
        (0, assert_1.default)(keychains[0].commonKeychain, 'Unable to find commonKeychain in keychains');
        const presign = {
            ...params,
            walletData: this._wallet,
            tssUtils: this.tssUtils,
            prv: this.getUserPrv(userPrvOptions),
            keychain: keychains[0],
            backupKeychain: keychains.length > 1 ? keychains[1] : null,
            bitgoKeychain: keychains.length > 2 ? keychains[2] : null,
            pub: keychains.map((k) => k.pub),
            reqId: params.reqId,
        };
        return this.signTypedDataTss(presign);
    }
    /**
     *  Sign a message using TSS
     * @param params
     * - Message
     * - custodianMessageId
     */
    async signMessage(params = {}) {
        if (!this.baseCoin.supportsMessageSigning()) {
            throw new Error(`Message signing not supported for ${this.baseCoin.getFullName()}`);
        }
        if (!params.message || !params.message.messageStandardType) {
            throw new Error('message and type required to sign message');
        }
        if (this._wallet.multisigType !== 'tss') {
            throw new Error('Message signing only supported for TSS wallets');
        }
        if (_.isFunction(this.baseCoin.encodeMessage)) {
            params.message.messageEncoded = this.baseCoin.encodeMessage(params.message.messageRaw);
        }
        const keychains = await this.baseCoin.keychains().getKeysForSigning({ wallet: this, reqId: params.reqId });
        const userPrvOptions = {
            ...params,
            keychain: keychains[0],
        };
        (0, assert_1.default)(keychains[0].commonKeychain, 'Unable to find commonKeychain in keychains');
        const presign = {
            ...params,
            walletData: this._wallet,
            tssUtils: this.tssUtils,
            prv: this.getUserPrv(userPrvOptions),
            keychain: keychains[0],
            backupKeychain: keychains.length > 1 ? keychains[1] : null,
            bitgoKeychain: keychains.length > 2 ? keychains[2] : null,
            reqId: params.reqId,
        };
        return this.signMessageTss(presign);
    }
    /**
     * Prepares and creates a sign message request for TSS wallets, that can be used later for signing.
     *
     * @param params - Parameters for creating the sign message request
     * @returns Promise<TxRequest> - The created transaction request for signing a message
     */
    async buildSignMessageRequest(params) {
        if (this._wallet.multisigType !== 'tss') {
            throw new Error('Message signing only supported for TSS wallets');
        }
        if (!this.baseCoin.supportsMessageSigning()) {
            throw new Error(`Message signing not supported for ${this.baseCoin.getFullName()}`);
        }
        if (!params.message?.messageRaw || !params.message?.messageStandardType) {
            throw new Error('message and type required to create message sign request');
        }
        const messageRaw = params.message.messageRaw;
        const messageStandardType = params.message.messageStandardType;
        const reqId = params.reqId || new utils_1.RequestTracer();
        try {
            const intentOption = {
                custodianMessageId: params.custodianMessageId,
                reqId,
                intentType: 'signMessage',
                isTss: true,
                messageRaw,
                messageStandardType,
                signerAddress: params.message.signerAddress,
            };
            if (!this.tssUtils) {
                throw new Error('TSS utilities not available for this wallet');
            }
            return await this.tssUtils.buildSignMessageRequest(intentOption);
        }
        catch (error) {
            throw new Error(`Failed to create message sign request: ${error}`);
        }
    }
    /**
     * Get the user private key from either a derivation or an encrypted keychain
     * @param [params.keychain / params.key] (object) or params.prv (string)
     * @param params.walletPassphrase (string)
     */
    getUserPrv(params = {}) {
        const userKeychain = params.keychain || params.key;
        let userPrv = params.prv;
        if (userPrv && typeof userPrv !== 'string') {
            throw new Error('prv must be a string');
        }
        // use the `derivedFromParentWithSeed` property from the user keychain as the `coldDerivationSeed`
        // if no other `coldDerivationSeed` was explicitly provided
        // Only for onchain multisig wallets, TSS key derivation happens during the signing process
        if (params.coldDerivationSeed === undefined &&
            params.keychain !== undefined &&
            params.keychain.derivedFromParentWithSeed !== undefined &&
            this.multisigType() === 'onchain') {
            params.coldDerivationSeed = params.keychain.derivedFromParentWithSeed;
        }
        if (userPrv && params.coldDerivationSeed) {
            // the derivation only makes sense when a key already exists
            const derivation = this.baseCoin.deriveKeyWithSeed({
                key: userPrv,
                seed: params.coldDerivationSeed,
            });
            userPrv = derivation.key;
        }
        else if (!userPrv) {
            if (!userKeychain || typeof userKeychain !== 'object') {
                throw new Error('keychain must be an object');
            }
            const userEncryptedPrv = userKeychain.encryptedPrv;
            if (!userEncryptedPrv) {
                throw new Error('keychain does not have property encryptedPrv');
            }
            if (!params.walletPassphrase) {
                throw new Error('walletPassphrase property missing');
            }
            userPrv = (0, keychain_1.decryptKeychainPrivateKey)(this.bitgo, userKeychain, params.walletPassphrase);
            if (!userPrv) {
                throw new Error('failed to decrypt user keychain');
            }
        }
        return userPrv;
    }
    /**
     * Get a transaction prebuild from BitGo, validate it, and then decrypt the user key and sign the transaction
     * @param params
     */
    async prebuildAndSignTransaction(params = {}) {
        if (params.eip1559 && params.gasPrice) {
            const error = new Error('Only one of params.eip1559 and params.gasPrice may be specified');
            error.code = 'both_gasPrice_and_eip1559gasModel_specified';
            throw error;
        }
        if (params.prebuildTx && params.recipients) {
            const error = new Error('Only one of prebuildTx and recipients may be specified');
            error.code = 'both_prebuildtx_and_recipients_specified';
            throw error;
        }
        if (params.recipients && !Array.isArray(params.recipients)) {
            const error = new Error('expecting recipients array');
            error.code = 'recipients_not_array';
            throw error;
        }
        if (_.isArray(this._permissions) && !this._permissions.includes('spend')) {
            const error = new Error('no spend permission on this wallet');
            error.code = 'user_not_allowed_to_spend_from_wallet';
            throw error;
        }
        if (params.receiveAddress && (params.type === 'transfer' || params.type === 'transferToken')) {
            const error = new Error(`cannot use receive address for TSS transactions of type ${params.type}`);
            error.code = 'receive_address_not_allowed_for_tss_withdrawals';
            throw error;
        }
        if (params.recipients && (params.type === 'fillNonce' || params.type === 'acceleration')) {
            const error = new Error(`cannot provide recipients for transaction type ${params.type}`);
            error.code = 'recipients_not_allowed_for_fillnonce_and_acceleration_tx_type';
            throw error;
        }
        if (params.apiVersion) {
            (0, txRequest_1.validateTxRequestApiVersion)(this, params.apiVersion);
        }
        // Doing a sanity check for password here to avoid doing further work if we know it's wrong
        const keychains = await this.getKeychainsAndValidatePassphrase({
            reqId: params.reqId,
            walletPassphrase: params.walletPassphrase,
            customSigningFunction: params.customSigningFunction,
        });
        let txPrebuildQuery;
        if (isPrebuildTransactionResult(params.prebuildTx) && params.prebuildTx.buildParams?.preview) {
            // If we prebuilt the txRequest with preview=true, then we should rebuild with preview=false to persist the request
            txPrebuildQuery = this.prebuildTransaction({
                ...params,
                ...{ ...params.prebuildTx.buildParams, preview: false },
            });
        }
        else {
            txPrebuildQuery = params.prebuildTx ? Promise.resolve(params.prebuildTx) : this.prebuildTransaction(params);
        }
        // the prebuild can be overridden by providing an explicit tx
        const txPrebuild = (await txPrebuildQuery);
        try {
            await this.baseCoin.verifyTransaction({
                txParams: { ...txPrebuild.buildParams, ...params },
                txPrebuild,
                wallet: this,
                verification: params.verification ?? {},
                reqId: params.reqId,
                walletType: this._wallet.multisigType,
            });
        }
        catch (e) {
            console.error('transaction prebuild failed local validation:', e.message);
            console.error('transaction params:', _.omit(params, ['keychain', 'prv', 'passphrase', 'walletPassphrase', 'key', 'wallet']));
            // Sanitize to remove _token from bitgo
            const sanitizedPrebuild = {
                ..._.omit(txPrebuild, ['wallet']),
                wallet: {
                    ...this,
                    bitgo: _.omit(this.bitgo, ['_token']),
                },
            };
            console.error('transaction prebuild:', sanitizedPrebuild);
            console.trace(e);
            throw e;
        }
        // pass our three keys
        const signingParams = {
            ...params,
            txPrebuild,
            wallet: this,
            keychain: keychains[0],
            backupKeychain: keychains.length > 1 ? keychains[1] : null,
            bitgoKeychain: keychains.length > 2 ? keychains[2] : null,
            reqId: params.reqId,
        };
        if (this._wallet.multisigType === 'onchain') {
            signingParams.pubs = keychains.map((k) => {
                (0, assert_1.default)(k.pub);
                return k.pub;
            });
        }
        if (signingParams.txPrebuild.txRequestId) {
            (0, assert_1.default)(this.tssUtils, 'tssUtils must be defined for TSS wallets');
            const txRequest = await this.tssUtils.getTxRequest(signingParams.txPrebuild.txRequestId, params.reqId);
            if (this.tssUtils.isPendingApprovalTxRequestFull(txRequest)) {
                return txRequest;
            }
        }
        try {
            return await this.signTransaction(signingParams);
        }
        catch (error) {
            if (error.message.includes('insufficient funds')) {
                error.code = 'insufficient_funds';
                error.walletBalances = {
                    balanceString: this.balanceString(),
                    confirmedBalanceString: this.confirmedBalanceString(),
                    spendableBalanceString: this.spendableBalanceString(),
                    balance: this.balance(),
                    confirmedBalance: this.confirmedBalance(),
                    spendableBalance: this.spendableBalance(),
                };
                error.txParams = {
                    ..._.omit(params, ['keychain', 'prv', 'passphrase', 'walletPassphrase', 'key', 'wallet']),
                    wallet: {
                        ...this,
                        bitgo: _.omit(this.bitgo, ['_token']),
                    },
                };
            }
            throw error;
        }
    }
    /**
     * Accelerate a transaction's confirmation using Child-Pays-For-Parent (CPFP) or Replace-By-Fee (RBF)
     * @param params
     */
    async accelerateTransaction(params = {}) {
        this.validateAccelerationParams(params);
        params.recipients = [];
        return await this.submitTransaction({
            ...(await this.prebuildAndSignTransaction(params)),
            ...BuildParams_1.BuildParams.encode(params),
        });
    }
    validateAccelerationParams(params) {
        if (!params.cpfpTxIds && !params.rbfTxIds) {
            const error = new Error('must pass cpfpTxIds or rbfTxIds');
            error.code = 'cpfptxids_or_rbftxids_required';
            throw error;
        }
        if (params.cpfpTxIds && params.rbfTxIds) {
            const error = new Error('cannot specify both cpfpTxIds and rbfTxIds');
            error.code = 'cannot_specify_both_cpfp_and_rbf_txids';
            throw error;
        }
        if (params.cpfpTxIds) {
            this.validateCpfpParams(params);
        }
        if (params.rbfTxIds) {
            this.validateRbfParams(params);
        }
        if (params.recipients !== undefined) {
            if (!Array.isArray(params.recipients) || params.recipients.length !== 0) {
                throw new Error(`invalid value for 'recipients': must be empty array when set`);
            }
        }
    }
    validateRbfParams(params) {
        if (!Array.isArray(params.rbfTxIds) || params.rbfTxIds.length !== 1) {
            const error = new Error('expecting rbfTxIds to be an array of length 1');
            error.code = 'rbftxids_not_array';
            throw error;
        }
        if (!params.feeMultiplier) {
            const error = new Error('feeMultiplier must be set');
            error.code = 'feemultiplier_not_set';
            throw error;
        }
        if (params.feeMultiplier <= 1) {
            const error = new Error('feeMultiplier must be a greater than 1');
            error.code = 'feemultiplier_greater_than_one';
            throw error;
        }
    }
    validateCpfpParams(params) {
        // TODO(BG-9349): change the last check to > 0 and the error message once platform allows multiple transactions to
        // be bumped in the same CPFP transaction
        if (!Array.isArray(params.cpfpTxIds) || params.cpfpTxIds.length !== 1) {
            const error = new Error('expecting cpfpTxIds to be an array of length 1');
            error.code = 'cpfptxids_not_array';
            throw error;
        }
        if (_.isUndefined(params.cpfpFeeRate)) {
            if (params.noCpfpFeeRate !== true) {
                const error = new Error('cpfpFeeRate must be set unless noCpfpFeeRate is set');
                error.code = 'cpfpfeerate_not_set';
                throw error;
            }
        }
        else {
            if (!_.isInteger(params.cpfpFeeRate) || params.cpfpFeeRate < 0) {
                const error = new Error('cpfpFeeRate must be a non-negative integer');
                error.code = 'cpfpfeerate_not_nonnegative_integer';
                throw error;
            }
        }
        if (_.isUndefined(params.maxFee)) {
            if (params.noMaxFee !== true) {
                const error = new Error('maxFee must be set unless noMaxFee is set');
                error.code = 'maxfee_not_set';
                throw error;
            }
        }
        else {
            if (!_.isInteger(params.maxFee) || params.maxFee < 0) {
                const error = new Error('maxFee must be a non-negative integer');
                error.code = 'maxfee_not_nonnegative_integer';
                throw error;
            }
        }
    }
    /**
     * Submit a half-signed transaction to BitGo
     * @param params
     * - txHex: transaction hex to submit
     * - halfSigned: object containing transaction (txHex or txBase64) to submit
     * @param reqId - request tracer request id
     */
    async submitTransaction(params = {}, reqId) {
        common.validateParams(params, [], ['otp', 'txHex', 'txRequestId']);
        const hasTxHex = !!params.txHex;
        const hasHalfSigned = !!params.halfSigned;
        if (params.txRequestId && (hasTxHex || hasHalfSigned)) {
            throw new Error('must supply exactly one of txRequestId, txHex, or halfSigned');
        }
        else if (!params.txRequestId && ((hasTxHex && hasHalfSigned) || (!hasTxHex && !hasHalfSigned))) {
            throw new Error('must supply either txHex or halfSigned, but not both');
        }
        return this.sendTransaction(params, reqId);
    }
    /**
     * Send coins to a recipient
     * @param params
     * @param params.address - the destination address
     * @param params.amount - the amount in satoshis/wei/base value to be sent
     * @param params.message - optional message to attach to transaction
     * @param params.data - [Ethereum Specific] optional data to pass to transaction
     * @param params.custodianTransactionId - [Ethereum/MMI Specific] id of transaction created via metamask
     * @param params.walletPassphrase - the passphrase to be used to decrypt the user key on this wallet
     * @param params.prv - the private key in string form, if walletPassphrase is not available
     * @param params.minConfirms - the minimum confirmation threshold for inputs
     * @param params.enforceMinConfirmsForChange - whether to enforce minConfirms for change inputs
     * @returns {*}
     */
    async send(params = {}) {
        common.validateParams(params, ['address'], ['message', 'data']);
        if (_.isUndefined(params.amount)) {
            throw new Error('missing required parameter amount');
        }
        if (_.isUndefined(params.address)) {
            throw new Error('missing required parameter address');
        }
        const coin = this.baseCoin;
        const amount = new bignumber_js_1.default(params.amount);
        const isAmountNegative = amount.isNegative();
        const isAmountZero = amount.isZero();
        const isAmountDecimal = !amount.isInteger();
        _.some([isAmountNegative, !coin.valuelessTransferAllowed() && isAmountZero, isAmountDecimal], (condition) => {
            if (condition) {
                throw new Error('invalid argument for amount - Integer greater than zero or numeric string expected');
            }
        });
        const recipients = [
            {
                address: params.address,
                amount: params.amount,
            },
        ];
        if (params.tokenName) {
            recipients[0].tokenName = params.tokenName;
        }
        if (params.data && coin.transactionDataAllowed()) {
            recipients[0].data = params.data;
        }
        const sendManyOptions = Object.assign({}, params, {
            recipients,
        });
        return this.sendMany(sendManyOptions);
    }
    /**
     * Send an ERC-721 NFT or ERC-1155 NFT(s).
     *
     * This function constructs the appropriate call data for an ERC-721/1155 token transfer,
     * and calls the token contract with the data, and amount 0. This transaction will always produce
     * a pending approval.
     *
     * @param sendOptions Options to specify how the transaction should be sent.
     * @param sendNftOptions Options to specify the NFT(s) to be sent.
     *
     * @return A pending approval for the transaction.
     */
    async sendNft(sendOptions, sendNftOptions) {
        const nftCollections = await this.getNftBalances();
        const { tokenContractAddress, recipientAddress, type } = sendNftOptions;
        const nftBalance = nftCollections.find((c) => c.metadata.tokenContractAddress === tokenContractAddress);
        if (!nftBalance) {
            throw new Error(`Collection not found for token contract ${tokenContractAddress}`);
        }
        if (!this.baseCoin.isValidAddress(recipientAddress)) {
            throw new Error(`Invalid recipient address ${recipientAddress}`);
        }
        const baseAddress = this.coinSpecific()?.baseAddress || this.coinSpecific()?.rootAddress;
        if (!baseAddress) {
            throw new Error('Missing base address for wallet');
        }
        if (nftBalance.type !== type) {
            throw new Error(`Specified NFT type ${type} does not match collection type ${nftBalance.type}`);
        }
        switch (sendNftOptions.type) {
            case 'ERC721': {
                if (!nftBalance.collections[sendNftOptions.tokenId]) {
                    throw new Error(`Token ${sendNftOptions.tokenId} not found in collection ${tokenContractAddress} or does not have a spendable balance`);
                }
                const data = this.baseCoin.buildNftTransferData({
                    ...sendNftOptions,
                    fromAddress: baseAddress,
                });
                let recipient;
                if (this.baseCoin.getFamily() === 'vet') {
                    recipient = {
                        address: recipientAddress,
                        amount: '1',
                        tokenData: data,
                    };
                }
                else {
                    recipient = {
                        address: sendNftOptions.tokenContractAddress,
                        amount: '0',
                        data,
                    };
                }
                return this.sendMany({
                    ...sendOptions,
                    recipients: [recipient],
                });
            }
            case 'ERC1155': {
                const entries = sendNftOptions.entries;
                for (const entry of entries) {
                    if (!nftBalance.collections[entry.tokenId]) {
                        throw new Error(`Token ${entry.tokenId} not found in collection ${sendNftOptions.tokenContractAddress} or does not have a spendable balance`);
                    }
                    if (nftBalance.collections[entry.tokenId] < entry.amount) {
                        throw new Error(`Amount ${entry.amount} exceeds spendable balance of ${nftBalance.collections[entry.tokenId]} for token ${entry.tokenId}`);
                    }
                }
                const data = this.baseCoin.buildNftTransferData({
                    ...sendNftOptions,
                    fromAddress: baseAddress,
                });
                return this.sendMany({
                    ...sendOptions,
                    recipients: [
                        {
                            address: sendNftOptions.tokenContractAddress,
                            amount: '0',
                            data: data,
                        },
                    ],
                });
            }
            case utils_1.TokenType.DIGITAL_ASSET: {
                if (!nftBalance.collections[sendNftOptions.tokenId]) {
                    throw new Error(`Token ${sendNftOptions.tokenId} not found in collection ${tokenContractAddress} or does not have a spendable balance`);
                }
                const tokenData = {
                    tokenType: sendNftOptions.type,
                    tokenQuantity: '1', // This NFT standard will always have quantity of 1
                    tokenContractAddress,
                    tokenId: sendNftOptions.tokenId,
                };
                return this.sendMany({
                    ...sendOptions,
                    recipients: [
                        {
                            address: recipientAddress,
                            amount: '1', // the amount needs to be non-zero for the transaction to be valid, it is ignored
                            tokenData,
                        },
                    ],
                });
            }
        }
    }
    /**
     * Send money to multiple recipients
     * 1. Gets the user keychain by checking the wallet for a key which has an encrypted prv
     * 2. Decrypts user key
     * 3. Creates the transaction with default fee
     * 4. Signs transaction with decrypted user key
     * 5. Sends the transaction to BitGo
     * @param {object} params
     * @param {{address: string, amount: string}} params.recipients - list of recipients and necessary recipient information
     * @param {Number} params.numBlocks - Estimates the approximate fee per kilobyte necessary for a transaction confirmation within numBlocks blocks
     * @param {Number} params.feeRate - the desired feeRate for the transaction in satothis/kB
     * @param {Number} params.maxFeeRate - upper limit for feeRate in satoshis/kB
     * @param {Number} params.minConfirms - all selected unspents will have at least this many confirmations
     * @param {Boolean} params.enforceMinConfirmsForChange - Enforces minConfirms on change inputs
     * @param {Number} params.targetWalletUnspents - The desired count of unspents in the wallet
     * @param {String} params.message - optional message to attach to transaction
     * @param {Number | String} params.minValue - Ignore unspents smaller than this amount of satoshis
     * @param {Number | String} params.maxValue - Ignore unspents larger than this amount of satoshis
     * @param {Number} params.sequenceId - The sequence ID of the transaction
     * @param {Number} params.lastLedgerSequence - Absolute max ledger the transaction should be accepted in, whereafter it will be rejected.
     * @param {Number} params.ledgerSequenceDelta - Relative ledger height (in relation to the current ledger) that the transaction should be accepted in, whereafter it will be rejected.
     * @param {Number} params.gasPrice - Custom gas price to be used for sending the transaction
     * @param {Boolean} params.noSplitChange - Set to true to disable automatic change splitting for purposes of unspent management
     * @param {Array} params.unspents - The unspents to use in the transaction. Each unspent should be in the form prevTxId:nOutput
     * @param {String} params.comment - Any additional comment to attach to the transaction
     * @param {String} params.otp - Two factor auth code to enable sending the transaction
     * @param {String} params.changeAddress - Specifies the destination of the change output
     * @param {Boolean} params.instant - Send this transaction using coin-specific instant sending method (if available)
     * @param {{value: String, type: String}} params.memo - Memo to use in transaction (supported by Stellar)
     * @param {String} params.type - Type of the transaction (e.g. trustline)
     * @param {{token: params, action: String, limit: String}[]} options.trustlines - Array of trustlines to manage (supported by Stellar)
     * @returns {*}
     */
    async sendMany(params = {}) {
        common.validateParams(params, [], ['comment', 'otp']);
        debug('sendMany called');
        const reqId = params.reqId || new utils_1.RequestTracer();
        params.reqId = reqId;
        this.bitgo.setRequestTracer(reqId);
        const coin = this.baseCoin;
        if (_.isObject(params.recipients)) {
            params.recipients.forEach(function (recipient) {
                coin.checkRecipient(recipient);
            });
        }
        if (this._wallet.multisigType === 'tss') {
            return this.sendManyTxRequests(params);
        }
        const selectParams = _.pick(params, [...this.prebuildWhitelistedParams(), 'comment', 'otp', 'hop']);
        if (this._wallet.type === 'custodial') {
            const extraParams = await this.baseCoin.getExtraPrebuildParams(Object.assign(params, { wallet: this }));
            Object.assign(selectParams, extraParams);
            return this.initiateTransaction(selectParams, reqId);
        }
        const halfSignedTransaction = await this.prebuildAndSignTransaction(params);
        const extraParams = await this.baseCoin.getExtraPrebuildParams(Object.assign(params, { wallet: this }));
        const finalTxParams = _.extend({}, halfSignedTransaction, selectParams, extraParams);
        return this.sendTransaction(finalTxParams, reqId);
    }
    /**
     * Recover an unsupported token from a BitGo multisig wallet
     * params are validated in Eth.prototype.recoverToken
     * @param params
     * @param params.tokenContractAddress the contract address of the unsupported token
     * @param params.recipient the destination address recovered tokens should be sent to
     * @param params.walletPassphrase the wallet passphrase
     * @param params.prv the xprv
     */
    async recoverToken(params = {}) {
        if (this.baseCoin.getFamily() !== 'eth') {
            throw new Error('token recovery only supported for eth wallets');
        }
        const { tokenContractAddress, recipient } = params;
        if (_.isUndefined(tokenContractAddress)) {
            throw new Error('missing required string parameter tokenContractAddress');
        }
        if (_.isUndefined(recipient)) {
            throw new Error('missing required string parameter recipient');
        }
        const recoverTokenOptions = Object.assign({ tokenContractAddress, recipient }, params, { wallet: this });
        return this.baseCoin.recoverToken(recoverTokenOptions);
    }
    /**
     * Get transaction metadata for the oldest transaction that is still pending or attempted
     * @param params
     * @returns {Object} Object with txid, walletId, tx, and fee (if supported for coin)
     */
    async getFirstPendingTransaction(params = {}) {
        return internal.getFirstPendingTransaction({ walletId: this.id() }, this.baseCoin, this.bitgo);
    }
    /**
     * Change the fee on the pending transaction that corresponds to the given txid to the given new fee
     * @param params
     * @param {String} params.txid The transaction Id corresponding to the transaction whose fee is to be changed
     * @param {String} [params.fee] Optional - The new fee to apply to the denoted transaction
     * @param {Object} [params.eip1559] Optional - the eip1559 values to apply to the denoted transaction
     * @returns {String} The transaction ID of the new transaction that contains the new fee rate
     */
    async changeFee(params = {}) {
        if (params.fee)
            common.validateParams(params, ['txid', 'fee'], []);
        if (params.eip1559)
            common.validateParams(params.eip1559, ['maxFeePerGas', 'maxPriorityFeePerGas']);
        return await this.bitgo
            .post(this.baseCoin.url('/wallet/' + this.id() + '/tx/changeFee'))
            .send(params)
            .result();
    }
    /**
     * Fetch info from merchant server
     * @param {Object} params The params passed into the function
     * @param {String} params.url The Url to retrieve info from
     * @returns {Object} The info returned from the merchant server
     * @deprecated
     */
    async getPaymentInfo(params = {}) {
        params = params || {};
        common.validateParams(params, ['url'], []);
        return await this.bitgo.get(this.url('/paymentInfo')).query(params).result();
    }
    /**
     * Send json payment response
     * @param {Object} params The params passed into the function
     * @param {String} params.paymentUrl - The url to send the fully signed transaction to
     * @param {String} params.txHex - The transaction hex of the payment
     * @param {String} params.memo {String} - A memo supplied by the merchant, to be inserted into the transfer as the comment
     * @param {String} params.expires {String} - ISO Date format of when the payment request expires
     * @returns {Object} The info returned from the merchant server Payment Ack
     * @deprecated
     */
    async sendPaymentResponse(params = {}) {
        return await this.bitgo.post(this.url('/sendPayment')).send(params).result();
    }
    /**
     * Create a policy rule
     * @param params
     * @param params.condition condition object
     * @param params.action action object
     * @returns {*}
     */
    async createPolicyRule(params = {}) {
        common.validateParams(params, ['id', 'type'], ['message']);
        if (!_.isObject(params.condition)) {
            throw new Error('missing parameter: conditions object');
        }
        if (!_.isObject(params.action)) {
            throw new Error('missing parameter: action object');
        }
        return await this.bitgo.post(this.url('/policy/rule')).send(params).result();
    }
    /**
     * Update a policy rule
     * @param params
     * @param params.condition condition object
     * @param params.action action object
     * @returns {*}
     */
    async setPolicyRule(params = {}) {
        common.validateParams(params, ['id', 'type'], ['message']);
        if (!_.isObject(params.condition)) {
            throw new Error('missing parameter: conditions object');
        }
        if (!_.isObject(params.action)) {
            throw new Error('missing parameter: action object');
        }
        return await this.bitgo.put(this.url('/policy/rule')).send(params).result();
    }
    /**
     * Remove Policy Rule
     * @param params
     * @returns {*}
     */
    async removePolicyRule(params = {}) {
        common.validateParams(params, ['id'], ['message']);
        return await this.bitgo.del(this.url('/policy/rule')).send(params).result();
    }
    /**
     * Remove this wallet
     * @param params
     * @returns {*}
     */
    async remove(params = {}) {
        return this.bitgo.del(this.url()).result();
    }
    /**
     * Fetches crossChain UTXOs
     * Currently only for AVAX
     * @param {string} params.sourceChain the sourcechain to pick UTXOs, if not given, then pick from all available chains [P, C]
     */
    fetchCrossChainUTXOs(params) {
        const query = _.pick(params, ['sourceChain']);
        return this.bitgo.get(this.url('/crossChainUnspents')).query(query).result();
    }
    /**
     * Extract a JSON representable version of this wallet
     */
    toJSON() {
        return this._wallet;
    }
    /**
     * Create a trading account from this wallet
     */
    toTradingAccount() {
        if (this.baseCoin.getFamily() !== 'ofc') {
            throw new Error('Can only convert an Offchain (OFC) wallet to a trading account');
        }
        return new trading_1.TradingAccount(this._wallet.enterprise, this, this.bitgo);
    }
    /**
     * Get the address book for this wallet
     */
    toAddressBook() {
        if (this.baseCoin.getFamily() !== 'ofc') {
            throw new Error('Can only use an Offchain (OFC) wallet for the address book');
        }
        return new address_book_1.AddressBook(this._wallet.enterprise, this.bitgo, this);
    }
    /**
     * Create a staking wallet from this wallet
     */
    toStakingWallet() {
        const isEthTss = this.baseCoin.getFamily() == 'eth' && this._wallet.coinSpecific?.walletVersion
            ? this._wallet.coinSpecific.walletVersion >= 3
            : false;
        return new staking_1.StakingWallet(this, isEthTss);
    }
    /**
     * Create a go staking wallet from this wallet
     */
    toGoStakingWallet() {
        if (this.baseCoin.getFamily() !== 'ofc') {
            throw new Error('Can only convert an Offchain (OFC) wallet to a staking wallet');
        }
        return new staking_1.GoStakingWallet(this);
    }
    /**
     * Creates and downloads PDF keycard for wallet (requires response from wallets.generateWallet)
     *
     * Note: this is example code and is not the version used on bitgo.com
     *
     * @param params
     *   * jsPDF - an instance of the jsPDF library
     *   * QRCode - an instance of the QRious library
     *   * userKeychain - a wallet's private user keychain
     *   * backupKeychain - a wallet's private backup keychain
     *   * bitgoKeychain - a wallet's private bitgo keychain
     *   * passphrase - the wallet passphrase
     *   * passcodeEncryptionCode - the encryption secret used for Box D
     *   * activationCode - a randomly generated six-digit activation code
     *   * walletKeyID - the Key ID used for deriving a cold wallet's signing key
     *   * backupKeyID - the Key ID used for deriving a cold wallet's backup key
     * @returns {*}
     */
    downloadKeycard(params = {}) {
        if (!window || !window.location) {
            throw new Error('The downloadKeycard function is only callable within a browser.');
        }
        // Grab parameters with default for activationCode
        const { jsPDF, QRCode, userKeychain, backupKeychain, bitgoKeychain, passphrase, passcodeEncryptionCode, walletKeyID, backupKeyID, activationCode = Math.floor(Math.random() * 900000 + 100000).toString(), } = params;
        if (!jsPDF || typeof jsPDF !== 'function') {
            throw new Error('Please pass in a valid jsPDF instance');
        }
        // Validate keychains
        if (!userKeychain || typeof userKeychain !== 'object') {
            throw new Error(`Wallet keychain must have a 'user' property`);
        }
        if (!backupKeychain || typeof backupKeychain !== 'object') {
            throw new Error('Backup keychain is required and must be an object');
        }
        if (!bitgoKeychain || typeof bitgoKeychain !== 'object') {
            throw new Error('Bitgo keychain is required and must be an object');
        }
        if (walletKeyID && typeof walletKeyID !== 'string') {
            throw new Error('walletKeyID must be a string');
        }
        if (backupKeyID && typeof backupKeyID !== 'string') {
            throw new Error('backupKeyID must be a string');
        }
        // Validate activation code if provided
        if (typeof activationCode !== 'string') {
            throw new Error('Activation Code must be a string');
        }
        if (activationCode.length !== 6) {
            throw new Error('Activation code must be six characters');
        }
        const coinShortName = this.baseCoin.type;
        const coinName = this.baseCoin.getFullName();
        const walletLabel = this._wallet.label;
        const doc = (0, internal_1.drawKeycard)({
            jsPDF,
            QRCode,
            encrypt: this.bitgo.encrypt,
            coinShortName,
            coinName,
            activationCode,
            walletLabel,
            passphrase,
            passcodeEncryptionCode,
            userKeychain,
            backupKeychain,
            bitgoKeychain,
            walletKeyID,
            backupKeyID,
        });
        // Save the PDF on the user's browser
        doc.save(`BitGo Keycard for ${walletLabel}.pdf`);
    }
    /**
     * Builds a set of consolidation transactions for a wallet.
     * @param params
     *     consolidateAddresses - these are the on-chain receive addresses we want to pick a consolidation amount from
     */
    async buildAccountConsolidations(params = {}) {
        if (!this.baseCoin.allowsAccountConsolidations()) {
            throw new Error(`${this.baseCoin.getFullName()} does not allow account consolidations.`);
        }
        // Whitelist params to build tx
        const whitelistedParams = _.pick(params, this.prebuildConsolidateAccountParams());
        debug('prebuilding consolidation transaction: %O', whitelistedParams);
        if (params.reqId) {
            this.bitgo.setRequestTracer(params.reqId);
        }
        // this could return 100 build transactions
        const buildResponse = (await this.bitgo
            .post(this.baseCoin.url('/wallet/' + this.id() + '/consolidateAccount/build'))
            .send(whitelistedParams)
            .result());
        if (buildResponse.length === 0) {
            throw new Error('No receive addresses with balance found to consolidate.');
        }
        // we need to step over each prebuild now - should be in an array in the body
        const consolidations = [];
        for (const consolidateAccountBuild of buildResponse) {
            let prebuild = (await this.baseCoin.postProcessPrebuild(Object.assign(consolidateAccountBuild, {
                wallet: this,
                buildParams: whitelistedParams,
            })));
            delete prebuild.wallet;
            delete prebuild.buildParams;
            prebuild = _.extend({}, prebuild, { walletId: this.id() });
            debug('final consolidation transaction prebuild: %O', prebuild);
            consolidations.push(prebuild);
        }
        return consolidations;
    }
    /**
     * Builds and sends a set of consolidation transactions for a wallet.
     * @param params
     *     prebuildTx   - this is the pre-build consolidation tx. this is a normally built tx with
     *                    an additional parameter of consolidateId.
     *     verification - normal keychains, etc. for verification
     */
    async sendAccountConsolidation(params = {}) {
        if (!this.baseCoin.allowsAccountConsolidations()) {
            throw new Error(`${this.baseCoin.getFullName()} does not allow account consolidations.`);
        }
        if (this._wallet.type === 'custodial' && this._wallet.multisigType !== 'tss') {
            params.type = 'consolidate';
            return this.initiateTransaction(params, params.reqId);
        }
        // one of a set of consolidation transactions
        if (typeof params.prebuildTx === 'string' || params.prebuildTx === undefined) {
            throw new Error('Invalid build of account consolidation.');
        }
        if (!params.prebuildTx.consolidateId) {
            throw new Error('Failed to find consolidation id on consolidation transaction.');
        }
        if (this._wallet.multisigType === 'tss') {
            if (!params.prebuildTx.txRequestId) {
                throw new Error('Consolidation request missing txRequestId.');
            }
            return await this.sendManyTxRequests(params);
        }
        const signedPrebuild = (await this.prebuildAndSignTransaction(params));
        // decorate with our consolidation id
        signedPrebuild.consolidateId = params.prebuildTx.consolidateId;
        delete signedPrebuild.wallet;
        return await this.submitTransaction(signedPrebuild, params.reqId);
    }
    /**
     * Builds and sends a set of account consolidations. This is intended to flush many balances to the root wallet balance.
     * @param params -
     *     consolidateAddresses - these are the on-chain receive addresses we want to pick a consolidation amount from
     */
    async sendAccountConsolidations(params = {}) {
        if (!this.baseCoin.allowsAccountConsolidations()) {
            throw new Error(`${this.baseCoin.getFullName()} does not allow account consolidations.`);
        }
        const apiVersion = params.apiVersion ??
            (this.tssUtils && this.tssUtils.supportedTxRequestVersions().includes('full') ? 'full' : undefined);
        // Doing a sanity check for password here to avoid doing further work if we know it's wrong
        await this.getKeychainsAndValidatePassphrase({
            reqId: params.reqId,
            walletPassphrase: params.walletPassphrase,
            customSigningFunction: params.customSigningFunction,
        });
        // this gives us a set of account consolidation transactions
        const unsignedBuilds = await this.buildAccountConsolidations({
            ...params,
            apiVersion: apiVersion,
        });
        if (unsignedBuilds && unsignedBuilds.length > 0) {
            // Get wallet's base address to validate destination addresses
            const baseAddress = this._wallet.coinSpecific?.baseAddress || this._wallet.coinSpecific?.rootAddress;
            // Validate all transactions
            for (const build of unsignedBuilds) {
                if (baseAddress) {
                    debug('verifying txHex', JSON.stringify(build));
                    //Verify transactions send funds to the base address
                    if (!(await this.baseCoin.verifyTransaction({
                        txPrebuild: build,
                        txParams: params,
                        verification: {
                            consolidationToBaseAddress: true,
                        },
                        wallet: this,
                        walletType: this._wallet.multisigType,
                    }))) {
                        throw new Error('Found output that does not consolidate funds to base address');
                    }
                }
            }
            const successfulTxs = [];
            const failedTxs = new Array();
            for (const unsignedBuild of unsignedBuilds) {
                // fold any of the parameters we used to build this transaction into the unsignedBuild
                const unsignedBuildWithOptions = Object.assign({}, params);
                unsignedBuildWithOptions.apiVersion = apiVersion;
                unsignedBuildWithOptions.prebuildTx = unsignedBuild;
                try {
                    const sendTx = await this.sendAccountConsolidation(unsignedBuildWithOptions);
                    successfulTxs.push(sendTx);
                }
                catch (e) {
                    failedTxs.push(e);
                }
            }
            return {
                success: successfulTxs,
                failure: failedTxs,
            };
        }
    }
    /**
     * Builds a set of transactions that enables the specified tokens
     * @param params -
     *    enableTokens: Token enablement operations we want to perform
     * @returns Unsigned transactions that enables the specified tokens
     */
    async buildTokenEnablements(params = { enableTokens: [] }) {
        const teConfig = this.baseCoin.getTokenEnablementConfig();
        if (!teConfig.requiresTokenEnablement) {
            throw new Error(`${this.baseCoin.getFullName()} does not require token enablements`);
        }
        // Validate wallet type if coin requires it
        if (typeof teConfig.validateWallet === 'function' && this._wallet.type) {
            teConfig.validateWallet(this._wallet.type);
        }
        if (params.enableTokens.length === 0) {
            throw new Error('No tokens are being specified');
        }
        if (params.recipients) {
            throw new Error('Can not specify recipients for token enablement transactions');
        }
        if (params.reqId) {
            this.bitgo.setRequestTracer(params.reqId);
        }
        // Split query if we can't enable multiple tokens in one tx
        if (!teConfig.supportsMultipleTokenEnablements && params.enableTokens.length > 1) {
            const queries = params.enableTokens.map(async (enableToken) => {
                return this.buildTokenEnablements({
                    ...params,
                    enableTokens: [enableToken],
                });
            });
            const results = await Promise.all(queries);
            return results.flat();
        }
        const buildParams = _.pick(params, this.prebuildWhitelistedParams());
        if (!buildParams.type) {
            buildParams.type = 'enabletoken';
        }
        // Check if we build with intent
        if (this._wallet.multisigType === 'tss') {
            return [await this.prebuildTransaction(buildParams)];
        }
        else {
            // Rewrite tokens into recipients for buildTransaction
            buildParams.recipients = params.enableTokens.map((token) => {
                // If token has non address, take the first wallet address is stored in its coin-specific property
                // In account-based coin implementations that use wallet contracts, the address is called baseAddress (e.g. eth-like, xtz)
                // for others it's called rootAddress (e.g. xrp, xlm, algo, trx)
                const address = token.address || this._wallet.coinSpecific?.baseAddress || this._wallet.coinSpecific?.rootAddress;
                if (!address) {
                    throw new Error('Wallet does not have base address, must specify with token param');
                }
                return {
                    tokenName: token.name,
                    address,
                    amount: '0',
                };
            });
            delete buildParams.enableTokens;
            const prebuildTx = await this.prebuildTransaction(buildParams);
            prebuildTx.buildParams = buildParams;
            return [prebuildTx];
        }
    }
    /**
     * Builds wallet initialization
     * @param {PrebuildTransactionOptions} params
     * @returns {Promise<PrebuildTransactionResult>}
     */
    async buildWalletInitialization(params) {
        if (!this.baseCoin.requiresWalletInitializationTransaction()) {
            throw new Error(`Wallet initialization is not required for ${this.baseCoin.getFullName()}`);
        }
        if (this._wallet.multisigType !== 'tss') {
            throw new Error('Wallet initialization transaction is only supported for TSS wallets');
        }
        if (params.reqId) {
            this.bitgo.setRequestTracer(params.reqId);
        }
        const buildParams = _.pick(params, this.prebuildWhitelistedParams());
        if (!buildParams.type) {
            buildParams.type = 'createAccount';
        }
        return await this.prebuildTransaction(buildParams);
    }
    /**
     * Signs and sends a single unsigned token enablement transaction
     * @param params
     * @returns
     *   - The response from sending the transaction for hot/cold wallets
     *   - The response from initiating the transaction for custodial wallets
     */
    async sendTokenEnablement(params = {}) {
        const teConfig = this.baseCoin.getTokenEnablementConfig();
        params.verification = params.verification ?? {};
        params.verification.verifyTokenEnablement = params.verification.verifyTokenEnablement ?? true;
        if (!teConfig.requiresTokenEnablement) {
            throw new Error(`${this.baseCoin.getFullName()} does not require token enablement transactions`);
        }
        // Validate wallet type if coin requires it
        if (teConfig.validateWallet && this._wallet.type) {
            teConfig.validateWallet(this._wallet.type);
        }
        if (typeof params.prebuildTx === 'string' || params.prebuildTx?.buildParams?.type !== 'enabletoken') {
            throw new Error('Invalid build of token enablement.');
        }
        if (this._wallet.multisigType === 'tss') {
            return await this.sendManyTxRequests(params);
        }
        else {
            switch (this._wallet.type) {
                case 'hot':
                case 'cold':
                    const signedPrebuild = await this.prebuildAndSignTransaction(params);
                    return await this.submitTransaction(signedPrebuild, params.reqId);
                case 'custodial':
                case 'backing':
                    return this.initiateTransaction(params.prebuildTx.buildParams, params.reqId);
            }
        }
    }
    /**
     * Some chains require tokens to be enabled before they can be received/sent.
     * This is a dedicated function that enables tokens.
     *
     * Builds, signs, and sends a set of transactions that enables the specified tokens
     * @param params -
     *    enableTokens: Token enablement operations we want to perform
     * @return
     *    success: Successful responses from sendTokenEnablement
     *    failure: Errors from failed transactions
     */
    async sendTokenEnablements(params = { enableTokens: [] }) {
        const unsignedBuilds = await this.buildTokenEnablements(params);
        const successfulTxs = [];
        const failedTxs = new Array();
        for (const unsignedBuild of unsignedBuilds) {
            const unsignedBuildWithOptions = {
                ...params,
                prebuildTx: unsignedBuild,
            };
            try {
                const sendTx = await this.sendTokenEnablement(unsignedBuildWithOptions);
                successfulTxs.push(sendTx);
            }
            catch (e) {
                failedTxs.push(e);
            }
        }
        return {
            success: successfulTxs,
            failure: failedTxs,
        };
    }
    /**
     * The chain canton, requires the wallet to be initialized by sending out a transaction
     * to be onboarded onto the validator.
     * Builds, Signs and sends a transaction that initializes the canton wallet
     * @param params
     */
    async sendWalletInitialization(params = {}) {
        const prebuildTx = await this.buildWalletInitialization(params);
        const unsignedBuildWithOptions = {
            ...params,
            prebuildTx,
        };
        const successfulTxs = [];
        const failedTxs = new Array();
        try {
            const sendTx = await this.sendManyTxRequests(unsignedBuildWithOptions);
            successfulTxs.push(sendTx);
        }
        catch (e) {
            failedTxs.push(e);
        }
        return {
            success: successfulTxs,
            failure: failedTxs,
        };
    }
    /* MARK: TSS Helpers */
    /**
     * Prebuilds a transaction for a TSS wallet.
     *
     * @param params prebuild transaction options
     */
    async prebuildTransactionTxRequests(params = {}) {
        const reqId = params.reqId || new utils_1.RequestTracer();
        this.bitgo.setRequestTracer(reqId);
        const apiVersion = (0, txRequest_1.getTxRequestApiVersion)(this, params.apiVersion);
        // Two options different implementations of fees seems to now be supported, for now we will support both to be backwards compatible
        // TODO(BG-59685): deprecate one of these so that we have a single way to pass fees
        let feeOptions;
        if (params.feeOptions) {
            feeOptions = params.feeOptions;
        }
        else if (params.gasPrice !== undefined || params.eip1559 !== undefined) {
            feeOptions =
                params.gasPrice !== undefined
                    ? { gasPrice: params.gasPrice, gasLimit: params.gasLimit }
                    : {
                        maxFeePerGas: Number(params.eip1559?.maxFeePerGas),
                        maxPriorityFeePerGas: Number(params.eip1559?.maxPriorityFeePerGas),
                        gasLimit: params.gasLimit,
                    };
        }
        else if (params.gasLimit !== undefined) {
            feeOptions = { gasLimit: params.gasLimit };
        }
        else {
            feeOptions = undefined;
        }
        let txRequest;
        switch (params.type) {
            case 'transfer':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'payment',
                    sequenceId: params.sequenceId,
                    comment: params.comment,
                    recipients: params.recipients || [],
                    memo: params.memo,
                    nonce: params.nonce,
                    feeOptions,
                    custodianTransactionId: params.custodianTransactionId,
                    unspents: params.unspents,
                    senderAddress: params.senderAddress,
                    isTestTransaction: params.isTestTransaction,
                }, apiVersion, params.preview);
                break;
            case 'transfertoken':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    isTss: params.isTss,
                    intentType: 'transferToken',
                    recipients: params.recipients || [],
                    nonce: params.nonce,
                    feeOptions,
                    unspents: params.unspents,
                    sequenceId: params.sequenceId,
                }, apiVersion, params.preview);
                break;
            case 'enabletoken':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'enableToken',
                    recipients: params.recipients || [],
                    enableTokens: params.enableTokens,
                    memo: params.memo,
                }, apiVersion, params.preview);
                break;
            case 'acceleration':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'acceleration',
                    comment: params.comment,
                    lowFeeTxid: params.lowFeeTxid,
                    receiveAddress: params.receiveAddress,
                    feeOptions,
                }, apiVersion, params.preview);
                break;
            case 'fillNonce':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'fillNonce',
                    comment: params.comment,
                    nonce: params.nonce,
                    receiveAddress: params.receiveAddress,
                    feeOptions,
                }, apiVersion, params.preview);
                break;
            case 'tokenApproval':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'tokenApproval',
                    tokenName: params.tokenName,
                }, apiVersion, params.preview);
                break;
            case 'createAccount':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'createAccount',
                    recipients: params.recipients || [],
                }, apiVersion, params.preview);
                break;
            case 'transferAccept': {
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'transferAccept',
                    txRequestId: params.txRequestId,
                    sequenceId: params.txRequestId,
                }, apiVersion, params.preview);
                break;
            }
            case 'transferReject': {
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'transferReject',
                    txRequestId: params.txRequestId,
                    sequenceId: params.txRequestId,
                }, apiVersion, params.preview);
                break;
            }
            case 'transferOfferWithdrawn': {
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'transferOfferWithdrawn',
                    transferOfferId: params.transferOfferId,
                    sequenceId: params.transferOfferId,
                }, apiVersion, params.preview);
                break;
            }
            case 'customTx':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'customTx',
                    sequenceId: params.sequenceId,
                    comment: params.comment,
                    solInstructions: params.solInstructions,
                    solVersionedTransactionData: params.solVersionedTransactionData,
                    aptosCustomTransactionParams: params.aptosCustomTransactionParams,
                    recipients: params.recipients || [],
                    nonce: params.nonce,
                    memo: params.memo,
                    feeOptions,
                }, apiVersion, params.preview);
                break;
            case 'bridgeFunds':
                txRequest = await this.tssUtils.prebuildTxWithIntent({
                    reqId,
                    intentType: 'bridgeFunds',
                    sequenceId: params.sequenceId,
                    comment: params.comment,
                    amount: params.intentAmount,
                    nonce: params.nonce,
                    feeOptions,
                }, apiVersion, params.preview);
                break;
            default:
                throw new Error(`transaction type not supported: ${params.type}`);
        }
        let unsignedTx;
        if (txRequest.apiVersion === 'full') {
            if (txRequest.transactions?.length !== 1) {
                throw new Error(`Expected a single unsigned tx for tx request with id: ${txRequest.txRequestId}`);
            }
            unsignedTx = txRequest.transactions[0].unsignedTx;
        }
        else {
            if (txRequest.unsignedTxs.length !== 1) {
                throw new Error(`Expected a single unsigned tx for tx request with id: ${txRequest.txRequestId}`);
            }
            unsignedTx = txRequest.unsignedTxs[0];
        }
        const whitelistedParams = _.pick(params, this.prebuildWhitelistedParams());
        return {
            walletId: this.id(),
            wallet: this,
            txRequestId: txRequest.txRequestId,
            txHex: unsignedTx.serializedTxHex,
            buildParams: whitelistedParams,
            feeInfo: unsignedTx.feeInfo,
            ...(txRequest.pendingApprovalId && {
                pendingApprovalId: txRequest.pendingApprovalId,
            }),
        };
    }
    /**
     * Signs a transaction from a TSS EdDSA wallet using external signer.
     *
     * @param params signing options
     */
    async signTransactionTssExternalSignerEdDSA(params = {}, coin) {
        let txRequestId = '';
        if (params.txRequestId) {
            txRequestId = params.txRequestId;
        }
        else if (params.txPrebuild && params.txPrebuild.txRequestId) {
            txRequestId = params.txPrebuild.txRequestId;
        }
        else {
            throw new Error('TxRequestId required to sign TSS transactions with External Signer.');
        }
        if (!params.customCommitmentGeneratingFunction) {
            throw new Error('Generator function for commitment required to sign transactions with External Signer.');
        }
        if (!params.customRShareGeneratingFunction) {
            throw new Error('Generator function for R share required to sign transactions with External Signer.');
        }
        if (!params.customGShareGeneratingFunction) {
            throw new Error('Generator function for G share required to sign transactions with External Signer.');
        }
        (0, assert_1.default)(this.tssUtils, 'tssUtils must be defined');
        // adding this to rebuild the transaction just before signing for EdDSA transaction using external signer
        const reqId = params.reqId || undefined;
        await this.tssUtils.deleteSignatureShares(txRequestId, reqId);
        try {
            const signedTxRequest = await this.tssUtils.signEddsaTssUsingExternalSigner(txRequestId, params.customCommitmentGeneratingFunction, params.customRShareGeneratingFunction, params.customGShareGeneratingFunction, reqId);
            return signedTxRequest;
        }
        catch (e) {
            throw new Error('failed to sign transaction ' + e);
        }
    }
    /**
     * Signs and sends a transaction request from a TSS (hot) wallet, or a SMC (cold) wallet with an external signer.
     * Meant to be used for a transaction request where the signing process is aborted.
     *
     * @param params
     *    txRequestId - The ID of the transaction request.
     *    walletPassphrase - The passphrase for the wallet.
     *    isTxRequestFull - Flag indicating if the transaction request is full.
     * @returns A promise that resolves to a SignedTransaction.
     */
    async signAndSendTxRequest(params) {
        if (params.isTxRequestFull) {
            await this.tssUtils?.deleteSignatureShares(params.txRequestId);
        }
        const ret = await this.getUserKeyAndSignTssTransaction({
            walletPassphrase: params.walletPassphrase,
            txRequestId: params.txRequestId,
        });
        if (!params.isTxRequestFull) {
            // It is assumed that if its not a full tx request, then it is a lite tx request
            const submitTx = await this.submitTransaction({
                txRequestId: params.txRequestId,
            });
            return submitTx;
        }
        return ret;
    }
    /**
     * Ensures signature shares are in a clean state before signing a transaction.
     * Automatically deletes signature shares for Full TxRequests before signing to prevent
     * issues with stale or partial signatures.
     * Use this for Express API and other integrations that need automatic cleanup of signatures.
     *
     * @param params - signing options
     * @returns signed transaction
     */
    async ensureCleanSigSharesAndSignTransaction(params = {}) {
        const txRequestId = params.txRequestId || params.txPrebuild?.txRequestId;
        if (txRequestId && this.tssUtils && this.multisigType() === 'tss') {
            const txRequest = await this.tssUtils.getTxRequest(txRequestId);
            // Always clean signature shares for Full TxRequests before signing
            const isFull = txRequest.apiVersion === 'full' &&
                ((txRequest.transactions ?? []).length > 0 || (txRequest.messages ?? []).length > 0);
            if (isFull) {
                await this.tssUtils.deleteSignatureShares(txRequestId);
            }
        }
        return this.signTransaction(params);
    }
    /**
     * Signs a transaction from a TSS ECDSA wallet using external signer.
     *
     * @param params signing options
     */
    async signTransactionTssExternalSignerECDSA(coin, params = {}) {
        let txRequestId = '';
        if (params.txRequestId) {
            txRequestId = params.txRequestId;
        }
        else if (params.txPrebuild && params.txPrebuild.txRequestId) {
            txRequestId = params.txPrebuild.txRequestId;
        }
        else {
            throw new Error('TxRequestId required to sign TSS transactions with External Signer.');
        }
        if (!params.customPaillierModulusGeneratingFunction) {
            throw new Error('Generator function for paillier modulus required to sign transactions with External Signer.');
        }
        if (!params.customKShareGeneratingFunction) {
            throw new Error('Generator function for K share required to sign transactions with External Signer.');
        }
        if (!params.customMuDeltaShareGeneratingFunction) {
            throw new Error('Generator function for MuDelta share required to sign transactions with External Signer.');
        }
        if (!params.customSShareGeneratingFunction) {
            throw new Error('Generator function for S share required to sign transactions with External Signer.');
        }
        try {
            (0, assert_1.default)(this.tssUtils, 'tssUtils must be defined');
            const signedTxRequest = await this.tssUtils.signEcdsaTssUsingExternalSigner({
                txRequest: txRequestId,
                reqId: params.reqId || new utils_1.RequestTracer(),
            }, utils_1.RequestType.tx, params.customPaillierModulusGeneratingFunction, params.customKShareGeneratingFunction, params.customMuDeltaShareGeneratingFunction, params.customSShareGeneratingFunction);
            return signedTxRequest;
        }
        catch (e) {
            throw new Error('failed to sign transaction ' + e);
        }
    }
    /**
     * Signs a transaction from a TSS ECDSA wallet using external signer.
     *
     * @param params signing options
     */
    async signTransactionTssExternalSignerECDSAMPCv2(coin, params = {}) {
        let txRequestId = '';
        if (params.txRequestId) {
            txRequestId = params.txRequestId;
        }
        else if (params.txPrebuild && params.txPrebuild.txRequestId) {
            txRequestId = params.txPrebuild.txRequestId;
        }
        else {
            throw new Error('TxRequestId required to sign TSS transactions with External Signer.');
        }
        if (!params.customMPCv2SigningRound1GenerationFunction) {
            throw new Error('Generator function for MPCv2 Round 1 share required to sign transactions with External Signer.');
        }
        if (!params.customMPCv2SigningRound2GenerationFunction) {
            throw new Error('Generator function for MPCv2 Round 2 share required to sign transactions with External Signer.');
        }
        if (!params.customMPCv2SigningRound3GenerationFunction) {
            throw new Error('Generator function for MPCv2 Round 3 share required to sign transactions with External Signer.');
        }
        try {
            (0, assert_1.default)(this.tssUtils, 'tssUtils must be defined');
            const signedTxRequest = await this.tssUtils.signEcdsaMPCv2TssUsingExternalSigner({
                txRequest: txRequestId,
                reqId: params.reqId || new utils_1.RequestTracer(),
            }, params.customMPCv2SigningRound1GenerationFunction, params.customMPCv2SigningRound2GenerationFunction, params.customMPCv2SigningRound3GenerationFunction);
            return signedTxRequest;
        }
        catch (e) {
            throw new Error('failed to sign transaction ' + e);
        }
    }
    /**
     * Signs a transaction from a TSS wallet.
     *
     * @param params signing options
     */
    async signTransactionTss(params = {}) {
        if (!params.txPrebuild) {
            throw new Error('txPrebuild required to sign transactions with TSS');
        }
        if (!params.txPrebuild.txRequestId) {
            throw new Error('txRequestId required to sign transactions with TSS');
        }
        if (!params.prv) {
            throw new Error('prv required to sign transactions with TSS');
        }
        try {
            return await this.tssUtils.signTxRequest({
                txRequest: params.txPrebuild.txRequestId,
                txParams: params.txPrebuild.buildParams,
                prv: params.prv,
                reqId: params.reqId || new utils_1.RequestTracer(),
                apiVersion: params.apiVersion,
            });
        }
        catch (e) {
            throw new Error('failed to sign transaction ' + e);
        }
    }
    /**
     * Signs a message from a TSS wallet.
     *
     * @param params signing options
     */
    async signMessageTss(params = {}) {
        if (!params.reqId) {
            params.reqId = new utils_1.RequestTracer();
        }
        if (!params.prv) {
            throw new Error('prv required to sign message with TSS');
        }
        try {
            let txRequest;
            (0, assert_1.default)(params.message, 'message required for message signing');
            const messageRaw = params.message.messageRaw;
            if (!params.message.txRequestId) {
                const intentOption = {
                    custodianMessageId: params.custodianMessageId,
                    reqId: params.reqId,
                    intentType: 'signMessage',
                    isTss: true,
                    messageRaw,
                    messageStandardType: params.message.messageStandardType,
                    signerAddress: params.message.signerAddress,
                };
                txRequest = await this.tssUtils.buildSignMessageRequest(intentOption);
                params.message.txRequestId = txRequest.txRequestId;
            }
            else {
                txRequest = await (0, tss_1.getTxRequest)(this.bitgo, this.id(), params.message.txRequestId, params.reqId);
            }
            (0, assert_1.default)(txRequest.messages && txRequest.messages.length > 0, 'Unable to find messages in txRequest for message signing');
            const messageEncoded = txRequest.messages[0].messageEncoded;
            const signedMessageRequest = await this.tssUtils.signTxRequestForMessage({
                txRequest,
                prv: params.prv,
                reqId: params.reqId || new utils_1.RequestTracer(),
                messageRaw,
                messageEncoded,
                bufferToSign: Buffer.from(messageEncoded, 'hex'),
            });
            (0, assert_1.default)(signedMessageRequest.messages, 'Unable to find messages in signedMessageRequest');
            if (this.baseCoin.getFamily() === statics_1.CoinFamily.ETH) {
                (0, assert_1.default)(signedMessageRequest.messages[0].combineSigShare, 'Unable to find combineSigShare in signedMessageRequest.messages');
            }
            (0, assert_1.default)(signedMessageRequest.messages[0].txHash, 'Unable to find txHash in signedMessageRequest.messages');
            return {
                coin: this.coin(),
                txHash: signedMessageRequest.messages[0].txHash,
                signature: signedMessageRequest.messages[0].txHash,
                messageRaw,
                messageEncoded,
                txRequestId: signedMessageRequest.txRequestId,
            };
        }
        catch (e) {
            throw new Error('failed to sign message ' + e);
        }
    }
    /**
     * Signs a typed data from a TSS wallet.
     * @param params
     * @private
     */
    async signTypedDataTss(params) {
        if (!params.reqId) {
            params.reqId = new utils_1.RequestTracer();
        }
        if (!params.prv) {
            throw new Error('prv required to sign typed data with TSS');
        }
        try {
            let txRequest;
            (0, assert_1.default)(params.typedData, 'typedData required for typed data signing');
            if (!params.typedData.txRequestId) {
                const intentOptions = {
                    custodianMessageId: params.custodianMessageId,
                    reqId: params.reqId,
                    intentType: 'signTypedStructuredData',
                    isTss: true,
                    typedDataRaw: params.typedData.typedDataRaw,
                    typedDataEncoded: params.typedData.typedDataEncoded.toString('hex'),
                };
                txRequest = await this.tssUtils.createTxRequestWithIntentForTypedDataSigning(intentOptions);
                params.typedData.txRequestId = txRequest.txRequestId;
            }
            else {
                txRequest = await (0, tss_1.getTxRequest)(this.bitgo, this.id(), params.typedData.txRequestId, params.reqId);
            }
            const signedTypedDataRequest = await this.tssUtils.signTxRequestForMessage({
                txRequest,
                prv: params.prv,
                reqId: params.reqId || new utils_1.RequestTracer(),
                messageRaw: JSON.stringify(params.typedData.typedDataRaw),
                messageEncoded: params.typedData.typedDataEncoded.toString('hex'),
                bufferToSign: params.typedData.typedDataEncoded,
            });
            (0, assert_1.default)(signedTypedDataRequest.messages, 'Unable to find messages in signedTypedDataRequest');
            (0, assert_1.default)(signedTypedDataRequest.messages[0].combineSigShare, 'Unable to find combineSigShare in signedTypedDataRequest.messages');
            (0, assert_1.default)(signedTypedDataRequest.messages[0].txHash, 'Unable to find txHash in signedTypedDataRequest.messages');
            return {
                coin: this.coin(),
                txHash: signedTypedDataRequest.messages[0].txHash,
                signature: signedTypedDataRequest.messages[0].txHash,
                messageRaw: params.typedData.typedDataRaw,
                messageEncoded: params.typedData.typedDataEncoded.toString('hex'),
                txRequestId: signedTypedDataRequest.txRequestId,
            };
        }
        catch (e) {
            throw new Error('failed to sign typed data ' + e);
        }
    }
    /**
     * Builds, signs, and sends a transaction from a TSS wallet.
     *
     * @param params send options
     */
    async sendManyTxRequests(params = {}) {
        params.apiVersion = (0, txRequest_1.getTxRequestApiVersion)(this, params.apiVersion);
        const signedTransaction = (await this.prebuildAndSignTransaction(params));
        if (!signedTransaction.txRequestId) {
            throw new Error('txRequestId missing from signed transaction');
        }
        if (params.apiVersion === 'full') {
            const latestTxRequest = await (0, tss_1.getTxRequest)(this.bitgo, this.id(), signedTransaction.txRequestId, params.reqId);
            const reqId = params.reqId || new utils_1.RequestTracer();
            this.bitgo.setRequestTracer(reqId);
            const transfer = await this.bitgo
                .post(this.bitgo.url('/wallet/' + this._wallet.id + '/txrequests/' + signedTransaction.txRequestId + '/transfers', 2))
                .send()
                .result();
            if (latestTxRequest.state === 'pendingApproval') {
                const pendingApprovals = new pendingApproval_1.PendingApprovals(this.bitgo, this.baseCoin);
                const pendingApproval = await pendingApprovals.get({
                    id: latestTxRequest.pendingApprovalId,
                });
                return {
                    pendingApproval: pendingApproval.toJSON(),
                    txRequest: latestTxRequest,
                };
            }
            return {
                transfer,
                txRequest: latestTxRequest,
                txid: (latestTxRequest.transactions ?? [])[0]?.signedTx?.id,
                tx: (latestTxRequest.transactions ?? [])[0]?.signedTx?.tx,
                status: transfer.state,
            };
        }
        const reqId = params.reqId || undefined;
        return this.tssUtils?.sendTxRequest(signedTransaction.txRequestId, reqId);
    }
    /**
     * Send funds from a fee address to a forwarder. Only supports eth-like coins.
     *
     * @param {Object} params - parameters object
     * @param {String} params.forwarderAddress - Address of the forwarder to send funds to.
     * @param {String} params.amount - Amount to send the forwarder (optional). If not given, defaults to sending an estimate of the amount needed for a fund recovery
     * @returns {*}
     */
    async fundForwarder(params) {
        if (_.isUndefined(params.forwarderAddress)) {
            throw new Error('forwarder address required');
        }
        const url = this.url('/fundForwarder');
        this._wallet = await this.bitgo.post(url).send(params).result();
        return this._wallet;
    }
    /**
     * Send funds from a fee address to a forwarder.
     *
     * @param {Object} params - parameters object
     * @param {List} params.forwarders - list of fund forwarder options
     * @returns {*}
     */
    async fundForwarders(params) {
        if (_.isUndefined(params.forwarders) || params.forwarders.length == 0) {
            throw new Error('fund forwarder options required');
        }
        const url = this.url('/fundforwarders');
        this._wallet = await this.bitgo.post(url).send(params).result();
        return this._wallet;
    }
    /**
     * Gets forwarder's balance
     * @param params - optional query parameters
     * @returns List of forwarder address and balance
     * if params is not set then returns low balance forwarders
     */
    async getForwarderBalance(params) {
        const query = {};
        if (params?.maximumBalance) {
            query.maximumBalance = params?.maximumBalance;
        }
        if (params?.minimumBalance) {
            query.minimumBalance = params?.minimumBalance;
        }
        const url = this.url(`/forwarders/balances`);
        const response = await this.bitgo.get(url).query(query).result();
        return response;
    }
    /**
     * Gets the ecdsa tss challenges for a wallet.
     * These are static challenges that have been verified by an enterprise admin.
     * Callers should verify that an enterprise admin signed the challenge values before using them.
     *
     * @returns {Promise<WalletEcdsaChallenges>}
     */
    async getChallengesForEcdsaSigning() {
        // note: this is not a coin specific route, we cannot use this.url(..)
        const url = this.bitgo.url(`/wallet/${this.id()}/challenges`, 2);
        return await this.bitgo.get(url).query({}).result();
    }
    sendTransaction(params, reqId) {
        // extract the whitelisted params from the top level, in case
        // other invalid params are present that would fail encoding
        // and fall back to the body params
        const whitelistedParams = this.baseCoin.preprocessBuildParams(_.pick(params, whitelistedSendParams));
        const reqTracer = reqId || new utils_1.RequestTracer();
        this.bitgo.setRequestTracer(reqTracer);
        return (0, postWithCodec_1.postWithCodec)(this.bitgo, this.baseCoin.url('/wallet/' + this.id() + '/tx/send'), t.intersection([public_types_1.TxSendBody, t.partial({ locktime: t.number })]), whitelistedParams).result();
    }
    initiateTransaction(params, reqId) {
        // extract the whitelisted params from the top level, in case
        // other invalid params are present that would fail encoding
        // and fall back to the body params
        const whitelistedParams = this.baseCoin.preprocessBuildParams(_.pick(params, whitelistedSendParams));
        const reqTracer = reqId || new utils_1.RequestTracer();
        this.bitgo.setRequestTracer(reqTracer);
        return (0, postWithCodec_1.postWithCodec)(this.bitgo, this.baseCoin.url('/wallet/' + this.id() + '/tx/initiate'), public_types_1.TxSendBody, whitelistedParams).result();
    }
    /**
     * Get wallet keychains and validate passphrase if necessary
     * @param {PrebuildTransactionOptions} params - prebuild transaction options
     * @param {string} params.walletPassphrase - wallet passphrase
     * @param {string} params.reqId - request id for tracing purposes
     * @param {Function} params.customSigningFunction - custom signing function for external signing
     * @returns {Promise<Keychain[]>}
     */
    async getKeychainsAndValidatePassphrase({ customSigningFunction, walletPassphrase, reqId, }) {
        const keychains = await this.baseCoin.keychains().getKeysForSigning({ wallet: this, reqId });
        // Doing a sanity check for password here to avoid doing further work if we know it's wrong
        // we ignore this check with if customSigningFunction is provided
        //  which means that the user is handling the signing in external signing mode
        if (!customSigningFunction && keychains?.[0]?.encryptedPrv && walletPassphrase) {
            if (!(0, keychain_1.decryptKeychainPrivateKey)(this.bitgo, keychains[0], walletPassphrase)) {
                const error = new Error(`unable to decrypt keychain with the given wallet passphrase`);
                error.code = 'wallet_passphrase_incorrect';
                throw error;
            }
        }
        return keychains;
    }
    /**
     * Approve token for use with a batcher contract
     * This function builds, signs, and sends a token approval transaction
     *
     * @param {string} walletPassphrase - The passphrase to be used to decrypt the user key
     * @param {string} tokenName - The name of the token to be approved
     * @returns {Promise<any>} The transaction details
     */
    async approveErc20Token(walletPassphrase, tokenName) {
        const reqId = new utils_1.RequestTracer();
        this.bitgo.setRequestTracer(reqId);
        let tokenApprovalBuild;
        try {
            const url = this.baseCoin.url(`/wallet/${this.id()}/token/approval/build`);
            tokenApprovalBuild = await this.bitgo
                .post(url)
                .send({
                tokenName: tokenName,
            })
                .result();
        }
        catch (e) {
            throw e;
        }
        const keychains = await this.getKeychainsAndValidatePassphrase({
            reqId,
            walletPassphrase,
        });
        const signingParams = {
            txPrebuild: tokenApprovalBuild,
            keychain: keychains[0],
            walletPassphrase,
            reqId,
        };
        const halfSignedTransaction = await this.signTransaction(signingParams);
        const finalTxParams = _.extend({}, halfSignedTransaction);
        return this.sendTransaction(finalTxParams, reqId);
    }
    /**
     * Build token approval transaction for ERC20 tokens
     * If walletPassphrase is provided, also signs and sends the transaction
     *
     * @param {string} tokenName - The name of the token to be approved
     * @param {string} [walletPassphrase] - Optional wallet passphrase for signing and sending
     * @returns {Promise<BuildTokenApprovalResponse | SubmitTransactionResponse>} The token approval build response or transaction details if signed
     */
    async buildErc20TokenApproval(tokenName, walletPassphrase) {
        const reqId = new utils_1.RequestTracer();
        this.bitgo.setRequestTracer(reqId);
        let tokenApprovalBuild;
        const url = this.baseCoin.url(`/wallet/${this.id()}/token/approval/build`);
        try {
            tokenApprovalBuild = await this.bitgo
                .post(url)
                .send({
                tokenName: tokenName,
            })
                .result();
        }
        catch (error) {
            throw new Error(`error building erc20 token approval tx: ${error}`);
        }
        if (!walletPassphrase) {
            return tokenApprovalBuild;
        }
        const keychains = await this.getKeychainsAndValidatePassphrase({
            reqId,
            walletPassphrase,
        });
        const signingParams = {
            txPrebuild: tokenApprovalBuild,
            keychain: keychains[0],
            walletPassphrase,
            reqId,
        };
        const halfSignedTransaction = await this.signTransaction(signingParams);
        const finalTxParams = _.extend({}, halfSignedTransaction);
        return this.sendTransaction(finalTxParams, reqId);
    }
}
exports.Wallet = Wallet;
