// @bitgo/sdk-coin-sui/dist/src/lib/mystenlab/types/coin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinSupply = exports.CoinBalance = exports.PaginatedCoins = exports.CoinStruct = void 0;
const superstruct_1 = require("superstruct");
const common_1 = require("./common");
exports.CoinStruct = (0, superstruct_1.object)({
    coinType: (0, superstruct_1.string)(),
    coinObjectId: common_1.ObjectId,
    version: (0, superstruct_1.number)(),
    digest: common_1.TransactionDigest,
    balance: (0, superstruct_1.number)(),
    lockedUntilEpoch: (0, superstruct_1.nullable)((0, superstruct_1.number)()),
    previousTransaction: common_1.TransactionDigest,
});
exports.PaginatedCoins = (0, superstruct_1.object)({
    data: (0, superstruct_1.array)(exports.CoinStruct),
    nextCursor: (0, superstruct_1.union)([common_1.ObjectId, (0, superstruct_1.literal)(null)]),
    hasNextPage: (0, superstruct_1.boolean)(),
});
exports.CoinBalance = (0, superstruct_1.object)({
    coinType: (0, superstruct_1.string)(),
    coinObjectCount: (0, superstruct_1.number)(),
    totalBalance: (0, superstruct_1.number)(),
    lockedBalance: (0, superstruct_1.object)({
        epochId: (0, superstruct_1.optional)((0, superstruct_1.number)()),
        number: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    }),
});
exports.CoinSupply = (0, superstruct_1.object)({
    value: (0, superstruct_1.number)(),
});
