// @bitgo/sdk-coin-sui/dist/src/lib/mystenlab/types/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuiSystemStateSummary = exports.SuiValidatorSummary = exports.CommitteeInfo = exports.DelegationStakingPool = exports.DelegationStakingPoolFields = exports.Contents = exports.ContentsFieldsWithdraw = exports.ContentsFields = exports.SuiSupplyFields = exports.StakeSubsidy = exports.StakeSubsidyFields = exports.DelegatedStake = exports.StakeObject = exports.Balance = void 0;
const superstruct_1 = require("superstruct");
const common_1 = require("./common");
const transactions_1 = require("./transactions");
// Staking
exports.Balance = (0, superstruct_1.object)({
    value: (0, superstruct_1.number)(),
});
exports.StakeObject = (0, superstruct_1.object)({
    stakedSuiId: common_1.ObjectId,
    stakeRequestEpoch: transactions_1.EpochId,
    stakeActiveEpoch: transactions_1.EpochId,
    principal: (0, superstruct_1.number)(),
    status: (0, superstruct_1.union)([(0, superstruct_1.literal)('Active'), (0, superstruct_1.literal)('Pending'), (0, superstruct_1.literal)('Unstaked')]),
    estimatedReward: (0, superstruct_1.optional)((0, superstruct_1.number)()),
});
exports.DelegatedStake = (0, superstruct_1.object)({
    validatorAddress: common_1.SuiAddress,
    stakingPool: common_1.ObjectId,
    stakes: (0, superstruct_1.array)(exports.StakeObject),
});
exports.StakeSubsidyFields = (0, superstruct_1.object)({
    balance: (0, superstruct_1.object)({ value: (0, superstruct_1.number)() }),
    distribution_counter: (0, superstruct_1.number)(),
    current_distribution_amount: (0, superstruct_1.number)(),
    stake_subsidy_period_length: (0, superstruct_1.number)(),
    stake_subsidy_decrease_rate: (0, superstruct_1.number)(),
});
exports.StakeSubsidy = (0, superstruct_1.object)({
    type: (0, superstruct_1.string)(),
    fields: exports.StakeSubsidyFields,
});
exports.SuiSupplyFields = (0, superstruct_1.object)({
    value: (0, superstruct_1.number)(),
});
exports.ContentsFields = (0, superstruct_1.object)({
    id: (0, superstruct_1.string)(),
    size: (0, superstruct_1.number)(),
    head: (0, superstruct_1.object)({ vec: (0, superstruct_1.array)() }),
    tail: (0, superstruct_1.object)({ vec: (0, superstruct_1.array)() }),
});
exports.ContentsFieldsWithdraw = (0, superstruct_1.object)({
    id: (0, superstruct_1.string)(),
    size: (0, superstruct_1.number)(),
});
exports.Contents = (0, superstruct_1.object)({
    type: (0, superstruct_1.string)(),
    fields: exports.ContentsFields,
});
exports.DelegationStakingPoolFields = (0, superstruct_1.object)({
    exchangeRates: (0, superstruct_1.object)({
        id: (0, superstruct_1.string)(),
        size: (0, superstruct_1.number)(),
    }),
    id: (0, superstruct_1.string)(),
    pendingStake: (0, superstruct_1.number)(),
    pendingPoolTokenWithdraw: (0, superstruct_1.number)(),
    pendingTotalSuiWithdraw: (0, superstruct_1.number)(),
    poolTokenBalance: (0, superstruct_1.number)(),
    rewardsPool: (0, superstruct_1.object)({ value: (0, superstruct_1.number)() }),
    activationEpoch: (0, superstruct_1.object)({ vec: (0, superstruct_1.array)() }),
    deactivationEpoch: (0, superstruct_1.object)({ vec: (0, superstruct_1.array)() }),
    suiBalance: (0, superstruct_1.number)(),
});
exports.DelegationStakingPool = (0, superstruct_1.object)({
    type: (0, superstruct_1.string)(),
    fields: exports.DelegationStakingPoolFields,
});
exports.CommitteeInfo = (0, superstruct_1.object)({
    epoch: (0, superstruct_1.number)(),
    /** Array of (validator public key, stake unit) tuple */
    validators: (0, superstruct_1.optional)((0, superstruct_1.array)((0, superstruct_1.tuple)([transactions_1.AuthorityName, (0, superstruct_1.number)()]))),
});
exports.SuiValidatorSummary = (0, superstruct_1.object)({
    suiAddress: common_1.SuiAddress,
    protocolPubkeyBytes: (0, superstruct_1.string)(),
    networkPubkeyBytes: (0, superstruct_1.string)(),
    workerPubkeyBytes: (0, superstruct_1.string)(),
    proofOfPossessionBytes: (0, superstruct_1.string)(),
    operationCapId: (0, superstruct_1.string)(),
    name: (0, superstruct_1.string)(),
    description: (0, superstruct_1.string)(),
    imageUrl: (0, superstruct_1.string)(),
    projectUrl: (0, superstruct_1.string)(),
    p2pAddress: (0, superstruct_1.string)(),
    netAddress: (0, superstruct_1.string)(),
    primaryAddress: (0, superstruct_1.string)(),
    workerAddress: (0, superstruct_1.string)(),
    nextEpochProtocolPubkeyBytes: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochProofOfPossession: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochNetworkPubkeyBytes: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochWorkerPubkeyBytes: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochNetAddress: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochP2pAddress: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochPrimaryAddress: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    nextEpochWorkerAddress: (0, superstruct_1.nullable)((0, superstruct_1.string)()),
    votingPower: (0, superstruct_1.number)(),
    gasPrice: (0, superstruct_1.number)(),
    commissionRate: (0, superstruct_1.number)(),
    nextEpochStake: (0, superstruct_1.number)(),
    nextEpochGasPrice: (0, superstruct_1.number)(),
    nextEpochCommissionRate: (0, superstruct_1.number)(),
    stakingPoolId: (0, superstruct_1.string)(),
    stakingPoolActivationEpoch: (0, superstruct_1.nullable)((0, superstruct_1.number)()),
    stakingPoolDeactivationEpoch: (0, superstruct_1.nullable)((0, superstruct_1.number)()),
    stakingPoolSuiBalance: (0, superstruct_1.number)(),
    rewardsPool: (0, superstruct_1.number)(),
    poolTokenBalance: (0, superstruct_1.number)(),
    pendingStake: (0, superstruct_1.number)(),
    pendingPoolTokenWithdraw: (0, superstruct_1.number)(),
    pendingTotalSuiWithdraw: (0, superstruct_1.number)(),
    exchangeRatesId: (0, superstruct_1.string)(),
    exchangeRatesSize: (0, superstruct_1.number)(),
});
exports.SuiSystemStateSummary = (0, superstruct_1.object)({
    epoch: (0, superstruct_1.number)(),
    protocolVersion: (0, superstruct_1.number)(),
    systemStateVersion: (0, superstruct_1.number)(),
    storageFundTotalObjectStorageRebates: (0, superstruct_1.number)(),
    storageFundNonRefundableBalance: (0, superstruct_1.number)(),
    referenceGasPrice: (0, superstruct_1.number)(),
    safeMode: (0, superstruct_1.boolean)(),
    safeModeStorageRewards: (0, superstruct_1.number)(),
    safeModeComputationRewards: (0, superstruct_1.number)(),
    safeModeStorageRebates: (0, superstruct_1.number)(),
    safeModeNonRefundableStorageFee: (0, superstruct_1.number)(),
    epochStartTimestampMs: (0, superstruct_1.number)(),
    epochDurationMs: (0, superstruct_1.number)(),
    stakeSubsidyStartEpoch: (0, superstruct_1.number)(),
    maxValidatorCount: (0, superstruct_1.number)(),
    minValidatorJoiningStake: (0, superstruct_1.number)(),
    validatorLowStakeThreshold: (0, superstruct_1.number)(),
    validatorVeryLowStakeThreshold: (0, superstruct_1.number)(),
    validatorLowStakeGracePeriod: (0, superstruct_1.number)(),
    stakeSubsidyBalance: (0, superstruct_1.number)(),
    stakeSubsidyDistributionCounter: (0, superstruct_1.number)(),
    stakeSubsidyCurrentDistributionAmount: (0, superstruct_1.number)(),
    stakeSubsidyPeriodLength: (0, superstruct_1.number)(),
    stakeSubsidyDecreaseRate: (0, superstruct_1.number)(),
    totalStake: (0, superstruct_1.number)(),
    activeValidators: (0, superstruct_1.array)(exports.SuiValidatorSummary),
    pendingActiveValidatorsId: (0, superstruct_1.string)(),
    pendingActiveValidatorsSize: (0, superstruct_1.number)(),
    pendingRemovals: (0, superstruct_1.array)((0, superstruct_1.number)()),
    stakingPoolMappingsId: (0, superstruct_1.string)(),
    stakingPoolMappingsSize: (0, superstruct_1.number)(),
    inactivePoolsId: (0, superstruct_1.string)(),
    inactivePoolsSize: (0, superstruct_1.number)(),
    validatorCandidatesId: (0, superstruct_1.string)(),
    validatorCandidatesSize: (0, superstruct_1.number)(),
    atRiskValidators: (0, superstruct_1.array)((0, superstruct_1.tuple)([common_1.SuiAddress, (0, superstruct_1.number)()])),
    validatorReportRecords: (0, superstruct_1.array)((0, superstruct_1.tuple)([common_1.SuiAddress, (0, superstruct_1.array)(common_1.SuiAddress)])),
});
