// bitgo/dist/src/bitgo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitGo = void 0;
//
// BitGo JavaScript SDK
//
// Copyright 2014, BitGo, Inc.  All Rights Reserved.
//
const pjson = require("../package.json");
const _ = require("lodash");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_api_1 = require("@bitgo/sdk-api");
const statics_1 = require("@bitgo/statics");
const coinFactory_1 = require("./v2/coinFactory");
const coins_1 = require("./v2/coins");
class BitGo extends sdk_api_1.BitGoAPI {
    /**
     * Constructor for BitGo Object
     */
    constructor(params = {}) {
        super(params);
        if (!sdk_core_1.common.validateParams(params, [], [
            'clientId',
            'clientSecret',
            'refreshToken',
            'accessToken',
            'userAgent',
            'customRootURI',
            'customBitcoinNetwork',
            'serverXpub',
            'stellarFederationServerUrl',
        ]) ||
            (params.useProduction && !_.isBoolean(params.useProduction)) ||
            (params.useAms && !_.isBoolean(params.useAms))) {
            throw new Error('invalid argument');
        }
        if (!params.clientId !== !params.clientSecret) {
            throw new Error('invalid argument - must provide both client id and secret');
        }
        this._useAms = !!params.useAms;
        this._version = pjson.version;
        this._userAgent = params.userAgent || 'BitGoJS/' + this.version();
        this._coinFactory = new sdk_core_1.CoinFactory();
    }
    /**
     * Initialize the coin factory with token configurations
     * @param tokenConfigMap - A map of token metadata from AMS
     */
    initCoinFactory(tokenConfigMap) {
        const coinMap = (0, statics_1.createTokenMapUsingTrimmedConfigDetails)(tokenConfigMap);
        this._coinFactory = new sdk_core_1.CoinFactory();
        (0, coinFactory_1.registerCoinConstructors)(this._coinFactory, coinMap);
    }
    /**
     * Fetch all the tokens and initialize the coin factory
     */
    async registerAllTokens() {
        if (!this._useAms) {
            throw new Error('registerAllTokens is only supported when useAms is set to true');
        }
        // Fetch mainnet assets for prod and adminProd environments, testnet assets for all other environments
        const assetEnvironment = ['prod', 'adminProd'].includes(this.getEnv()) ? 'mainnet' : 'testnet';
        const url = this.url(`/assets/list/${assetEnvironment}`);
        const tokenConfigMap = (await this.executeAssetRequest(url));
        this.initCoinFactory(tokenConfigMap);
    }
    /**
     * Create a basecoin object
     * @param coinName
     */
    coin(coinName) {
        if (this._useAms) {
            return this._coinFactory.getInstance(this, coinName);
        }
        return coinFactory_1.GlobalCoinFactory.getInstance(this, coinName);
    }
    /**
     * Register a token in the coin factory
     * @param tokenConfig - The token metadata from AMS
     */
    async registerToken(tokenName) {
        if (!this._useAms) {
            throw new Error('registerToken is only supported when useAms is set to true');
        }
        //do not register a coin/token if it's already registered
        if (this._coinFactory.hasCoin(tokenName)) {
            return;
        }
        //ofc is not present in statics coin map
        if (tokenName === 'ofc') {
            this._coinFactory.register(tokenName, coins_1.Ofc.createInstance);
            return;
        }
        // Get the coin/token details only if it's not present in statics library
        let staticsBaseCoin;
        if (statics_1.coins.has(tokenName)) {
            staticsBaseCoin = statics_1.coins.get(tokenName);
        }
        else {
            const url = this.url(`/assets/name/${tokenName}`);
            const tokenConfig = (await this.executeAssetRequest(url));
            staticsBaseCoin = (0, statics_1.createToken)(tokenConfig);
        }
        if (!staticsBaseCoin) {
            throw new sdk_core_1.UnsupportedCoinError(tokenName);
        }
        if (staticsBaseCoin.isToken) {
            const formattedTokenConfig = (0, statics_1.getFormattedTokenConfigForCoin)(staticsBaseCoin);
            if (!formattedTokenConfig) {
                throw new sdk_core_1.UnsupportedCoinError(tokenName);
            }
            const tokenConstructor = (0, coinFactory_1.getTokenConstructor)(formattedTokenConfig);
            if (!tokenConstructor) {
                throw new sdk_core_1.UnsupportedCoinError(tokenName);
            }
            this._coinFactory.registerToken(staticsBaseCoin, tokenConstructor);
        }
        else {
            const coinConstructor = (0, coinFactory_1.getCoinConstructor)(tokenName);
            if (!coinConstructor) {
                throw new sdk_core_1.UnsupportedCoinError(tokenName);
            }
            this._coinFactory.registerToken(staticsBaseCoin, coinConstructor);
        }
    }
    /**
     * Create a basecoin object for a virtual token
     * @param tokenName
     */
    async token(tokenName) {
        await this.fetchConstants();
        return this.coin(tokenName);
    }
}
exports.BitGo = BitGo;
