// bitgo/dist/src/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokens = exports.testnetTokens = exports.mainnetTokens = exports.ethGasConfigs = exports.defaults = exports.defaultConstants = void 0;
const statics_1 = require("@bitgo/statics");
Object.defineProperty(exports, "ethGasConfigs", { enumerable: true, get: function () { return statics_1.ethGasConfigs; } });
Object.defineProperty(exports, "mainnetTokens", { enumerable: true, get: function () { return statics_1.mainnetTokens; } });
Object.defineProperty(exports, "testnetTokens", { enumerable: true, get: function () { return statics_1.testnetTokens; } });
Object.defineProperty(exports, "tokens", { enumerable: true, get: function () { return statics_1.tokens; } });
const sdk_core_1 = require("@bitgo/sdk-core");
Object.defineProperty(exports, "defaults", { enumerable: true, get: function () { return sdk_core_1.defaults; } });
Object.defineProperty(exports, "defaultConstants", { enumerable: true, get: function () { return sdk_core_1.defaultConstants; } });
