// bitgo/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalCoinFactory = exports.Environments = exports.Buffer = exports.sjcl = exports.bitcoin = void 0;
exports.setNetwork = setNetwork;
exports.getNetwork = getNetwork;
exports.getNetworkObj = getNetworkObj;
/**
 * @prettier
 */
//
// index.js - Module definition for BitGoJS
//
// Copyright 2022, BitGo, Inc.  All Rights Reserved.
//
const _ = require("lodash");
const sdk_core_1 = require("@bitgo/sdk-core");
__exportStar(require("@bitgo/sdk-api"), exports);
const utxolib = require("@bitgo/utxo-lib");
const coinFactory_1 = require("./v2/coinFactory");
Object.defineProperty(exports, "GlobalCoinFactory", { enumerable: true, get: function () { return coinFactory_1.GlobalCoinFactory; } });
__exportStar(require("./bitgo"), exports);
// Expose legacy "bitcoin" API (mostly HDNode)
/** @deprecated */
var sdk_core_2 = require("@bitgo/sdk-core");
Object.defineProperty(exports, "bitcoin", { enumerable: true, get: function () { return sdk_core_2.bitcoin; } });
/** @deprecated */
exports.sjcl = require('@bitgo/sjcl');
var buffer_1 = require("buffer");
Object.defineProperty(exports, "Buffer", { enumerable: true, get: function () { return buffer_1.Buffer; } });
exports.Environments = _.cloneDeep(sdk_core_1.common.Environments);
__exportStar(require("@bitgo/sdk-core"), exports);
__exportStar(require("./v2"), exports);
/**
 * Set the network, i.e. either "bitcoin" for production with real bitcoin, or
 * "testnet" for development with testnet bitcoin.
 *
 * @deprecated
 */
function setNetwork(network) {
    sdk_core_1.common.setNetwork(network);
}
/**
 * Get the network. Returns either "bitcoin" or "testnet".
 *
 * @deprecated
 */
function getNetwork() {
    return sdk_core_1.common.getNetwork();
}
/**
 * @deprecated
 */
function getNetworkObj() {
    return utxolib.networks[sdk_core_1.common.getNetwork()];
}
setNetwork('testnet');
