// @bitgo/abstract-utxo/dist/cjs/src/transaction/bip322.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeBIP322BroadcastableMessage = serializeBIP322BroadcastableMessage;
exports.deserializeBIP322BroadcastableMessage = deserializeBIP322BroadcastableMessage;
exports.verifyTransactionFromBroadcastableMessage = verifyTransactionFromBroadcastableMessage;
exports.generateBIP322MessageListAndVerifyFromMessageBroadcastable = generateBIP322MessageListAndVerifyFromMessageBroadcastable;
const sdk_core_1 = require("@bitgo/sdk-core");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const t = __importStar(require("io-ts"));
const BIP322MessageInfo = t.type({
    address: t.string,
    message: t.string,
    pubkeys: t.array(t.string),
    scriptType: t.union([
        t.literal('p2sh'),
        t.literal('p2shP2wsh'),
        t.literal('p2wsh'),
        t.literal('p2tr'),
        t.literal('p2trMusig2'),
    ]),
});
const BIP322MessageBroadcastable = t.type({
    txHex: t.string,
    messageInfo: t.array(BIP322MessageInfo),
});
function assertPubkeyTriple(pubkeys) {
    if (pubkeys.length !== 3) {
        throw new Error(`Expected exactly 3 pubkeys, got ${pubkeys.length}`);
    }
    return pubkeys;
}
function serializeBIP322BroadcastableMessage(message) {
    return Buffer.from(JSON.stringify(message), 'utf8').toString('hex');
}
function deserializeBIP322BroadcastableMessage(hex) {
    const json = JSON.parse(Buffer.from(hex, 'hex').toString('utf8'));
    return (0, sdk_core_1.decodeOrElse)(BIP322MessageBroadcastable.name, BIP322MessageBroadcastable, json, (error) => {
        throw new Error(`Failed to decode ${BIP322MessageBroadcastable.name}: ${error}`);
    });
}
function verifyTransactionFromBroadcastableMessage(message, coinName) {
    if (coinName !== 'btc' && coinName !== 'tbtc4') {
        throw new Error('Only tbtc4 or btc coinNames are supported.');
    }
    const network = coinName;
    if ((0, wasm_utxo_1.hasPsbtMagic)(Buffer.from(message.txHex, 'hex'))) {
        const psbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(Buffer.from(message.txHex, 'hex'), network);
        try {
            message.messageInfo.forEach((info, inputIndex) => {
                wasm_utxo_1.bip322.verifyBip322PsbtInputWithPubkeys(psbt, inputIndex, {
                    message: info.message,
                    pubkeys: assertPubkeyTriple(info.pubkeys),
                    scriptType: info.scriptType,
                });
            });
            return true;
        }
        catch (error) {
            return false;
        }
    }
    else {
        const tx = wasm_utxo_1.Transaction.fromBytes(Buffer.from(message.txHex, 'hex'));
        try {
            message.messageInfo.forEach((info, inputIndex) => {
                wasm_utxo_1.bip322.verifyBip322TxInputWithPubkeys(tx, inputIndex, {
                    message: info.message,
                    pubkeys: assertPubkeyTriple(info.pubkeys),
                    scriptType: info.scriptType,
                });
            });
            return true;
        }
        catch (error) {
            return false;
        }
    }
}
function generateBIP322MessageListAndVerifyFromMessageBroadcastable(messageBroadcastables, coinName) {
    // Map from the address to the message. If there are duplicates of the address, make sure that the
    // message is the same. If there are duplicate addresses and the messages are not the same, throw an error.
    const addressMap = new Map();
    messageBroadcastables.forEach((message, index) => {
        if (verifyTransactionFromBroadcastableMessage(message, coinName)) {
            message.messageInfo.forEach((info) => {
                const { address, message: msg } = info;
                if (addressMap.has(address)) {
                    if (addressMap.get(address) !== msg) {
                        throw new Error(`Duplicate address ${address} has different messages`);
                    }
                }
                else {
                    addressMap.set(address, msg);
                }
            });
        }
        else {
            throw new Error(`Message Broadcastable ${index} did not have a successful validation`);
        }
    });
    return Array.from(addressMap.entries()).map(([address, message]) => ({
        address,
        message,
    }));
}
