// @bitgo/abstract-utxo/dist/cjs/src/transaction/recipient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isScriptRecipient = isScriptRecipient;
exports.fromExtendedAddressFormat = fromExtendedAddressFormat;
exports.fromExtendedAddressFormatToScript = fromExtendedAddressFormatToScript;
exports.toOutputScript = toOutputScript;
exports.toExtendedAddressFormat = toExtendedAddressFormat;
exports.assertValidTransactionRecipient = assertValidTransactionRecipient;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const ScriptRecipientPrefix = 'scriptPubKey:';
/**
 * Check if the address is a script recipient (starts with `scriptPubKey:`).
 * @param address
 */
function isScriptRecipient(address) {
    return address.toLowerCase().startsWith(ScriptRecipientPrefix.toLowerCase());
}
/**
 * An extended address is one that encodes either a regular address or a hex encoded script with the prefix `scriptPubKey:`.
 * This function converts the extended address format to either a script or an address.
 * @param extendedAddress
 */
function fromExtendedAddressFormat(extendedAddress) {
    if (isScriptRecipient(extendedAddress)) {
        return { script: extendedAddress.slice(ScriptRecipientPrefix.length) };
    }
    return { address: extendedAddress };
}
function fromExtendedAddressFormatToScript(extendedAddress, coinName) {
    const result = fromExtendedAddressFormat(extendedAddress);
    if ('script' in result) {
        return Buffer.from(result.script, 'hex');
    }
    return Buffer.from(wasm_utxo_1.address.toOutputScriptWithCoin(result.address, coinName));
}
function toOutputScript(v, coinName) {
    if (typeof v === 'string') {
        return fromExtendedAddressFormatToScript(v, coinName);
    }
    if ('script' in v) {
        return Buffer.from(v.script, 'hex');
    }
    if ('address' in v) {
        return fromExtendedAddressFormatToScript(v.address, coinName);
    }
    throw new Error('invalid input');
}
const OP_RETURN = 0x6a;
/**
 * Convert a script or address to the extended address format.
 * @param script
 * @param coinName
 * @returns if the script is an OP_RETURN script, then it will be prefixed with `scriptPubKey:`, otherwise it will be converted to an address.
 */
function toExtendedAddressFormat(script, coinName) {
    return script[0] === OP_RETURN
        ? `${ScriptRecipientPrefix}${script.toString('hex')}`
        : wasm_utxo_1.address.fromOutputScriptWithCoin(script, coinName);
}
function assertValidTransactionRecipient(output) {
    // In the case that this is an OP_RETURN output or another non-encodable scriptPubkey, we dont have an address.
    // We will verify that the amount is zero, and if it isnt then we will throw an error.
    if (!output.address || isScriptRecipient(output.address)) {
        if (output.amount.toString() !== '0') {
            throw new Error(`Only zero amounts allowed for non-encodeable scriptPubkeys: amount: ${output.amount}, address: ${output.address}`);
        }
    }
}
