// @bitgo/abstract-utxo/dist/cjs/src/transaction/signTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.signTransaction = signTransaction;
const lodash_1 = __importDefault(require("lodash"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const debug_1 = __importDefault(require("debug"));
const descriptor_1 = require("../descriptor");
const keychains_1 = require("../keychains");
const wasmUtil_1 = require("../wasmUtil");
const fixedScript = __importStar(require("./fixedScript"));
const descriptor = __importStar(require("./descriptor"));
const decode_1 = require("./decode");
const debug = (0, debug_1.default)('bitgo:abstract-utxo:transaction:signTransaction');
function getSignerKeychain(userPrv) {
    if (userPrv === undefined) {
        return undefined;
    }
    if (typeof userPrv !== 'string') {
        throw new Error('expected user private key to be a string');
    }
    const signerKeychain = wasm_utxo_1.BIP32.fromBase58(userPrv);
    if (signerKeychain.isNeutered()) {
        throw new Error('expected user private key but received public key');
    }
    debug(`Here is the public key of the xprv you used to sign: ${signerKeychain.neutered().toBase58()}`);
    return signerKeychain;
}
async function signTransaction(coin, bitgo, params) {
    const txPrebuild = params.txPrebuild;
    if (lodash_1.default.isUndefined(txPrebuild) || !lodash_1.default.isObject(txPrebuild)) {
        if (!lodash_1.default.isUndefined(txPrebuild) && !lodash_1.default.isObject(txPrebuild)) {
            throw new Error(`txPrebuild must be an object, got type ${typeof txPrebuild}`);
        }
        throw new Error('missing txPrebuild parameter');
    }
    const tx = coin.decodeTransactionFromPrebuild(params.txPrebuild);
    const signerKeychain = getSignerKeychain(params.prv);
    const { wallet } = params;
    if (wallet && (0, descriptor_1.isDescriptorWallet)(wallet)) {
        if (!signerKeychain) {
            throw new Error('missing signer');
        }
        if (!(0, wasmUtil_1.isUtxoLibPsbt)(tx) && !(tx instanceof Uint8Array)) {
            throw new Error('descriptor wallets require PSBT format transactions');
        }
        const walletKeys = (0, keychains_1.toBip32Triple)(await (0, keychains_1.fetchKeychains)(coin, wallet));
        const descriptorMap = (0, descriptor_1.getDescriptorMapFromWallet)(wallet, walletKeys, (0, descriptor_1.getPolicyForEnv)(bitgo.env));
        const psbt = (0, wasmUtil_1.toWasmPsbt)(tx);
        descriptor.signPsbt(psbt, descriptorMap, signerKeychain, {
            onUnknownInput: 'throw',
        });
        return { txHex: Buffer.from(psbt.serialize()).toString('hex') };
    }
    else {
        const signedTx = await fixedScript.signTransaction(coin, tx, getSignerKeychain(params.prv), coin.name, {
            walletId: params.txPrebuild.walletId,
            txInfo: params.txPrebuild.txInfo,
            isLastSignature: params.isLastSignature ?? false,
            signingStep: params.signingStep,
            allowNonSegwitSigningWithoutPrevTx: params.allowNonSegwitSigningWithoutPrevTx ?? false,
            pubs: params.pubs,
            cosignerPub: params.cosignerPub,
        });
        const buffer = Buffer.isBuffer(signedTx) ? signedTx : (0, decode_1.encodeTransaction)(signedTx);
        return { txHex: buffer.toString('hex') };
    }
}
