// @bitgo/sdk-coin-oas/dist/test/unit/oas.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should = __importStar(require("should"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const index_1 = require("../../src/index");
const resources_1 = require("../resources");
const nock_1 = __importDefault(require("nock"));
const sdk_core_1 = require("@bitgo/sdk-core");
const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
describe('OASYS chain', function () {
    before(function () {
        bitgo.safeRegister('oas', index_1.Oas.createInstance);
        bitgo.safeRegister('toas', index_1.Toas.createInstance);
        bitgo.initializeTestVars();
    });
    describe('Basic Coin Info', function () {
        it('should return the right info for oas', function () {
            const oas = bitgo.coin('oas');
            oas.should.be.an.instanceof(index_1.Oas);
            oas.getChain().should.equal('oas');
            oas.getFamily().should.equal('oas');
            oas.getFullName().should.equal('Oasys');
            oas.getBaseFactor().should.equal(1e18);
            oas.supportsTss().should.equal(true);
            oas.allowsAccountConsolidations().should.equal(false);
        });
        it('should return the right info for toas', function () {
            const toas = bitgo.coin('toas');
            toas.should.be.an.instanceof(index_1.Toas);
            toas.getChain().should.equal('toas');
            toas.getFamily().should.equal('oas');
            toas.getFullName().should.equal('Testnet Oasys');
            toas.getBaseFactor().should.equal(1e18);
            toas.supportsTss().should.equal(true);
            toas.allowsAccountConsolidations().should.equal(false);
        });
    });
});
describe('Build Unsigned Sweep for Self-Custody Cold Wallets - (MPCv2)', function () {
    const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
    const explorerUrl = sdk_core_1.common.Environments[bitgo.getEnv()].oasExplorerBaseUrl;
    const maxFeePerGasvalue = 20000000000;
    const maxPriorityFeePerGasValue = 10000000000;
    const chain_id = 9372;
    const gasLimitvalue = 500000;
    it('should generate an unsigned sweep without derivation path', async () => {
        (0, nock_1.default)(explorerUrl)
            .get('/api')
            .twice()
            .query(resources_1.mockDataUnsignedSweep.getTxListRequest)
            .reply(200, resources_1.mockDataUnsignedSweep.getTxListResponse);
        (0, nock_1.default)(explorerUrl)
            .get('/api')
            .query(resources_1.mockDataUnsignedSweep.getBalanceRequest)
            .reply(200, resources_1.mockDataUnsignedSweep.getBalanceResponse);
        const baseCoin = bitgo.coin('toas');
        const transaction = (await baseCoin.recover({
            userKey: resources_1.mockDataUnsignedSweep.userKey,
            backupKey: resources_1.mockDataUnsignedSweep.backupKey,
            walletContractAddress: resources_1.mockDataUnsignedSweep.walletBaseAddress,
            recoveryDestination: resources_1.mockDataUnsignedSweep.recoveryDestination,
            isTss: true,
            eip1559: { maxFeePerGas: maxFeePerGasvalue, maxPriorityFeePerGas: maxPriorityFeePerGasValue },
            gasLimit: gasLimitvalue,
            replayProtectionOptions: {
                chain: chain_id,
                hardfork: 'london',
            },
        }));
        should.exist(transaction);
        transaction.should.have.property('txRequests');
        transaction.txRequests.length.should.equal(1);
        const txRequest = transaction.txRequests[0];
        txRequest.should.have.property('walletCoin');
        txRequest.walletCoin.should.equal('toas');
        txRequest.should.have.property('transactions');
        txRequest.transactions.length.should.equal(1);
        const tx = txRequest.transactions[0];
        tx.should.have.property('nonce');
        tx.should.have.property('unsignedTx');
        tx.unsignedTx.should.have.property('serializedTxHex');
        tx.unsignedTx.should.have.property('signableHex');
        tx.unsignedTx.should.have.property('derivationPath');
        tx.unsignedTx.should.have.property('feeInfo');
        tx.unsignedTx.feeInfo?.should.have.property('fee');
        tx.unsignedTx.feeInfo?.should.have.property('feeString');
        tx.unsignedTx.should.have.property('parsedTx');
        tx.unsignedTx.parsedTx?.should.have.property('spendAmount');
        tx.unsignedTx.parsedTx?.should.have.property('outputs');
    });
});
