// @bitgo/public-types/dist/src/utils/primitives.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionalUnknownObject = exports.optionalUnknown = exports.optionalStringOrNumber = exports.optionalNumber = exports.optionalBoolean = exports.optionalStringArray = exports.optionalString = void 0;
const t = __importStar(require("io-ts"));
const optional_1 = require("./optional");
exports.optionalString = (0, optional_1.Optional)(t.string);
exports.optionalStringArray = (0, optional_1.Optional)(t.array(t.string));
exports.optionalBoolean = (0, optional_1.Optional)(t.boolean);
exports.optionalNumber = (0, optional_1.Optional)(t.number);
exports.optionalStringOrNumber = (0, optional_1.Optional)(t.union([t.string, t.number]));
exports.optionalUnknown = (0, optional_1.Optional)(t.intersection([t.unknown, t.partial({})]));
exports.optionalUnknownObject = (0, optional_1.Optional)(t.record(t.string, t.unknown));
