// @bitgo/sdk-coin-sui/dist/test/unit/transferTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const statics_1 = require("@bitgo/statics");
const should_1 = __importDefault(require("should"));
const src_1 = require("../../src");
const testData = __importStar(require("../resources/sui"));
describe('Sui Transfer Transaction', () => {
    let tx;
    const config = statics_1.coins.get('tsui');
    beforeEach(() => {
        tx = new src_1.TransferTransaction(config);
    });
    describe('Empty transaction', () => {
        it('should throw empty transaction', function () {
            should_1.default.throws(() => tx.toJson(), 'Empty transaction');
            should_1.default.throws(() => tx.toBroadcastFormat(), 'Empty transaction');
        });
    });
    describe('From raw transaction', () => {
        it('should build a transfer from raw hex', function () {
            tx.fromRawTransaction(testData.TRANSFER);
            const json = tx.toJson();
            should_1.default.equal(json.sender, testData.sender.address);
        });
        it('should fail to build a transfer from incorrect raw hex', function () {
            should_1.default.throws(() => tx.fromRawTransaction('random' + testData.TRANSFER), 'incorrect raw data');
        });
    });
    describe('Explain transaction', () => {
        it('should explain a transfer pay transaction', function () {
            tx.fromRawTransaction(testData.TRANSFER);
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: 'BxoeGXbBCuw6VFEcgwHHUAKrCoAsGanPB39kdVVKZZcR',
                outputs: [
                    {
                        address: testData.recipients[0].address,
                        amount: testData.recipients[0].amount,
                    },
                    {
                        address: testData.recipients[1].address,
                        amount: testData.recipients[1].amount,
                    },
                ],
                outputAmount: `${testData.AMOUNT * 2}`,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: testData.gasData.budget.toString() },
                type: 0,
            });
        });
        it('should fail to explain transaction with invalid raw hex', function () {
            should_1.default.throws(() => tx.fromRawTransaction('randomString'), 'Invalid transaction');
        });
    });
});
