// @bitgo/utxo-core/dist/esm/descriptor/Output.js
import assert from 'assert';
import { Descriptor } from '@bitgo/wasm-utxo';
import { getFixedOutputSum } from '../Output';
import { getDescriptorAtIndexCheckScript } from './derive';
export function isInternalOutput(output) {
    return 'descriptor' in output && output.descriptor !== undefined;
}
export function isExternalOutput(output) {
    return !isInternalOutput(output);
}
/**
 * @return the sum of the external outputs that are not 'max'
 * @param outputs
 */
export function getExternalFixedAmount(outputs) {
    return getFixedOutputSum(outputs.filter(isExternalOutput));
}
export function toDerivedDescriptorWalletOutput(output, descriptorMap) {
    const descriptor = descriptorMap.get(output.descriptorName);
    if (!descriptor) {
        throw new Error(`Descriptor not found: ${output.descriptorName}`);
    }
    assert(descriptor instanceof Descriptor);
    const descriptorAtIndex = getDescriptorAtIndexCheckScript(descriptor, output.descriptorIndex, output.witnessUtxo.script, output.descriptorName);
    return {
        hash: output.hash,
        index: output.index,
        witnessUtxo: output.witnessUtxo,
        descriptor: descriptorAtIndex,
    };
}
