// @bitgo/sdk-coin-sui/dist/test/local_fullnode/RpcClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcClient = exports.RpcError = void 0;
const util_1 = __importDefault(require("util"));
const axios_1 = __importDefault(require("axios"));
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('RpcClient');
function elideResponse(method) {
    return method === 'sui_getProtocolConfig';
}
function unwrapResult(method, v) {
    if ('error' in v) {
        debug('< %s ERROR', method, v.error);
        throw new Error(JSON.stringify(v.error));
    }
    if (elideResponse(method)) {
        debug('< %s ...', method);
    }
    else {
        debug('< %s', method, util_1.default.inspect(v.result, { depth: 10 }));
    }
    return v.result;
}
class RpcError extends Error {
    constructor(rpcError) {
        super(`RPC error: ${rpcError.message} (code=${rpcError.code})`);
        this.rpcError = rpcError;
    }
    static isRpcErrorWithCode(e, code) {
        return e instanceof RpcError && e.rpcError.code === code;
    }
}
exports.RpcError = RpcError;
/** Wrapper around https://docs.sui.io/sui-jsonrpc */
class RpcClient {
    constructor(url) {
        // Running counter, increments every request
        this.id = 0;
        this.url = url;
    }
    static async createCheckedConnection(url) {
        const rpcClient = new RpcClient(url);
        const { protocolVersion } = await rpcClient.exec('sui_getProtocolConfig');
        const chainId = await rpcClient.exec('sui_getChainIdentifier');
        debug('Connected to', url, 'protocolVersion', protocolVersion, 'chainId', chainId);
        return rpcClient;
    }
    async exec(method, ...params) {
        params = params.filter((v) => v !== undefined);
        try {
            debug('>', this.url, method, params);
            const response = await axios_1.default.post(this.url, {
                jsonrpc: '2.0',
                method,
                params,
                id: `${this.id++}`,
            });
            return unwrapResult(method, response.data);
        }
        catch (e) {
            if (e.isAxiosError && e.response) {
                e = e;
                debug('< %s ERROR', method, e.response.statusText, util_1.default.inspect(e.response.data, { depth: 10 }));
                e = e;
                const { error = {} } = e.response.data;
                throw new RpcError(error);
            }
            throw e;
        }
    }
    /**
     * https://docs.sui.io/sui-jsonrpc#suix_getCoins
     */
    async getCoins(owner, coinType, cursor, limit) {
        return this.exec('suix_getCoins', owner, coinType, cursor, limit);
    }
    async executeTransactionBlock(tx_bytes, signatures, options, request_type) {
        return this.exec('sui_executeTransactionBlock', tx_bytes, signatures, {
            showInput: true,
            showRawInput: true,
            showEffects: true,
            showEvents: true,
            showObjectChanges: true,
            showBalanceChanges: true,
        }, request_type);
    }
    /**
     * https://docs.sui.io/sui-jsonrpc#suix_getValidatorsApy
     * APY = Annual Percentage Yield
     */
    async getValidatorsApy() {
        return this.exec('suix_getValidatorsApy');
    }
    /**
     * https://docs.sui.io/sui-jsonrpc#suix_getStakes
     */
    async getStakes(owner) {
        return this.exec('suix_getStakes', owner);
    }
    /**
     * https://docs.sui.io/sui-jsonrpc#sui_getObject
     */
    async getObject(object_id) {
        return this.exec('sui_getObject', object_id, { showData: true });
    }
}
exports.RpcClient = RpcClient;
