// @bitgo/sdk-coin-sui/dist/test/local_fullnode/transactions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
process.env.DEBUG = 'RpcClient,SuiTransactionTypes';
const assert_1 = __importDefault(require("assert"));
const util_1 = __importDefault(require("util"));
const faucet_1 = require("./faucet");
const debug_1 = __importDefault(require("debug"));
const crypto_1 = require("crypto");
const src_1 = require("../../src");
const RpcClient_1 = require("./RpcClient");
const getBuilderFactory_1 = require("../unit/getBuilderFactory");
const iface_1 = require("../../src/lib/iface");
const constants_1 = require("../../src/lib/constants");
const debug = (0, debug_1.default)('SuiTransactionTypes');
async function getAllCoins(conn, address) {
    return (await conn.getCoins(address)).data.map((v) => {
        return {
            digest: v.digest,
            objectId: v.coinObjectId,
            version: v.version,
        };
    });
}
/**
 * @returns The stakes array with the stakedSui amount reduced by amount. If amount is undefined, the stakedSui is removed.
 */
function subtractStake(stakes, stakedSui, amount) {
    return stakes.flatMap((s) => {
        if (s.stakedSuiId === stakedSui.stakedSuiId) {
            if (amount === undefined) {
                // remove the stake
                return [];
            }
            else {
                // reduce the stake by amount
                return [{ ...s, principal: Number(s.principal) - amount }];
            }
        }
        else {
            // keep the stake unchanged
            return [s];
        }
    });
}
/**
 * Asserts that the stakesAfter array is the same as the stakesBefore array with the stakedSui amount reduced by amount.
 */
function assertReducedStake(stakesBefore, stakesAfter, stakedSui, amount) {
    /*
     * Normalize the stake objects by converting the principal to a number and removing the estimatedReward.
     */
    function normStake(s) {
        return { ...s, principal: Number(s.principal), estimatedReward: undefined };
    }
    stakesBefore = stakesBefore.map(normStake);
    stakesAfter = stakesAfter.map(normStake);
    assert_1.default.deepStrictEqual(stakesAfter, subtractStake(stakesBefore, stakedSui, amount));
}
async function getStakes(conn, owner, params = {}) {
    const all = await conn.getStakes(owner);
    const result = await Promise.all(all.flatMap((s) => s.stakes
        .filter((v) => params.filterStatus === undefined || v.status === params.filterStatus)
        .filter((v) => params.filterMinValue === undefined || params.filterMinValue <= Number(v.principal))));
    if (result.length) {
        return result;
    }
    const { attempts = 60, sleepMs = 1000 } = params;
    if (0 < attempts) {
        await new Promise((resolve) => setTimeout(resolve, sleepMs));
        return await getStakes(conn, owner, { ...params, attempts: attempts - 1, sleepMs });
    }
    else {
        throw new Error('getAllActiveStakedSuis: no active staked suis found');
    }
}
async function resolveStakedSui(conn, stake) {
    return (await conn.getObject(stake.stakedSuiId)).data;
}
function getKeyPair(seed) {
    const seedBuf = (0, crypto_1.createHash)('sha256').update(seed).digest();
    return new src_1.KeyPair({ seed: seedBuf });
}
async function signAndSubmit(conn, keyPair, txb) {
    txb.sign({ key: keyPair.getKeys().prv });
    const tx = (await txb.build());
    debug('tx', util_1.default.inspect(tx.suiTransaction.tx, { depth: 10 }));
    const result = await conn.executeTransactionBlock(tx.toBroadcastFormat(), [
        Buffer.from(tx.serializedSig).toString('base64'),
    ]);
    if (result.effects?.status.status !== 'success') {
        throw new Error(`Transaction failed: ${JSON.stringify(result.effects?.status)}`);
    }
}
async function fundFromFaucet(url, v, amount = 100e9) {
    if (typeof v !== 'string') {
        v = v.getAddress();
    }
    await new faucet_1.Faucet(url).getCoins(v, 10e9);
}
describe('Sui Transaction Types', function () {
    if (process.env.DRONE) {
        console.log('skipping local_fullnode/transactions.ts on drone');
        return;
    }
    const keyPair = getKeyPair('test');
    const address = keyPair.getAddress();
    debug('address', address);
    const fullnodeUrl = process.env.SUI_FULLNODE_URL || 'http://127.0.0.1:9000';
    const faucetUrl = process.env.SUI_FAUCET_URL || 'http://127.0.0.1:9123';
    async function getDefaultGasData(keyPair) {
        return {
            owner: keyPair.getAddress(),
            payment: await getAllCoins(conn, keyPair.getAddress()),
            budget: 100000000,
            price: constants_1.DUMMY_SUI_GAS_PRICE,
        };
    }
    let conn;
    let validator;
    before('establish connection', async function () {
        conn = await RpcClient_1.RpcClient.createCheckedConnection(fullnodeUrl);
        const { apys } = await conn.getValidatorsApy();
        validator = apys[0].address;
    });
    before('fund via faucet', async function () {
        if (faucetUrl) {
            await fundFromFaucet(faucetUrl, address);
        }
    });
    it('has coins', async function () {
        const { data } = await conn.getCoins(address);
        assert_1.default.notStrictEqual(data.length, 0);
    });
    it('can transfer coins', async function () {
        const builder = (0, getBuilderFactory_1.getBuilderFactory)('tsui').getTransferBuilder();
        const txb = builder
            .type(iface_1.SuiTransactionType.Transfer)
            .sender(address)
            .send([{ address, amount: (111111).toString() }])
            .gasData(await getDefaultGasData(keyPair));
        await signAndSubmit(conn, keyPair, txb);
    });
    async function stakeAmount(keyPair, amount) {
        const builder = (0, getBuilderFactory_1.getBuilderFactory)('tsui').getStakingBuilder();
        const txb = builder
            .type(iface_1.SuiTransactionType.AddStake)
            .sender(keyPair.getAddress())
            .stake([{ amount, validatorAddress: validator }])
            .gasData(await getDefaultGasData(keyPair));
        await signAndSubmit(conn, keyPair, txb);
    }
    it('can stake coins', async function () {
        await stakeAmount(keyPair, 1e9);
    });
    function testUnstake(amount) {
        describe(`unstake (amount=${amount})`, function () {
            const keyPairStakeTest = getKeyPair(`stake-test-amount-${amount !== undefined}`);
            const address = keyPairStakeTest.getAddress();
            before('stake coins', async function () {
                await fundFromFaucet(faucetUrl, address);
                await stakeAmount(keyPairStakeTest, 10e9);
            });
            it(`can unstake`, async function () {
                const activeStakedSui = await getStakes(conn, address, {
                    filterStatus: 'Active',
                    filterMinValue: constants_1.MIN_STAKING_THRESHOLD + (amount || 0),
                });
                (0, assert_1.default)(activeStakedSui.length > 0, 'No staked coins found');
                for (const stakedSui of activeStakedSui.slice(0, 3)) {
                    debug('unstaking', stakedSui);
                    const builder = (0, getBuilderFactory_1.getBuilderFactory)('tsui').getUnstakingBuilder();
                    const stakedBefore = await getStakes(conn, address);
                    const txb = builder
                        .type(iface_1.SuiTransactionType.WithdrawStake)
                        .sender(address)
                        .unstake({ stakedSui: await resolveStakedSui(conn, stakedSui), amount })
                        .gasData(await getDefaultGasData(keyPairStakeTest));
                    await signAndSubmit(conn, keyPairStakeTest, txb);
                    assertReducedStake(stakedBefore, await getStakes(conn, address), stakedSui, amount);
                }
            });
        });
    }
    testUnstake(undefined);
    testUnstake(1e9);
});
