// @bitgo/wasm-utxo/dist/esm/js/ecpair.js
import { WasmECPair } from "./wasm/wasm_utxo.js";
/**
 * ECPair wrapper class for elliptic curve key pair operations
 */
export class ECPair {
    _wasm;
    constructor(_wasm) {
        this._wasm = _wasm;
    }
    /**
     * Create an ECPair instance from a WasmECPair instance (internal use)
     * @internal
     */
    static fromWasm(wasm) {
        return new ECPair(wasm);
    }
    /**
     * Convert ECPairArg to ECPair instance
     * @param key - The ECPair key in various formats
     * @returns ECPair instance
     */
    static from(key) {
        // Short-circuit if already an ECPair instance
        if (key instanceof ECPair) {
            return key;
        }
        // If it's a WasmECPair instance, wrap it
        if (key instanceof WasmECPair) {
            return new ECPair(key);
        }
        // Parse from Buffer/Uint8Array
        // Check length to determine if it's a private key (32 bytes) or public key (33 bytes)
        if (key.length === 32) {
            const wasm = WasmECPair.from_private_key(key);
            return new ECPair(wasm);
        }
        else if (key.length === 33) {
            const wasm = WasmECPair.from_public_key(key);
            return new ECPair(wasm);
        }
        else {
            throw new Error(`Invalid key length: ${key.length}. Expected 32 bytes (private key) or 33 bytes (compressed public key)`);
        }
    }
    /**
     * Create an ECPair from a private key (always uses compressed keys)
     * @param buffer - The 32-byte private key
     * @returns An ECPair instance
     */
    static fromPrivateKey(buffer) {
        const wasm = WasmECPair.from_private_key(buffer);
        return new ECPair(wasm);
    }
    /**
     * Create an ECPair from a compressed public key
     * @param buffer - The compressed public key bytes (33 bytes)
     * @returns An ECPair instance
     */
    static fromPublicKey(buffer) {
        const wasm = WasmECPair.from_public_key(buffer);
        return new ECPair(wasm);
    }
    /**
     * Create an ECPair from a WIF string (auto-detects network from WIF)
     * @param wifString - The WIF-encoded private key string
     * @returns An ECPair instance
     */
    static fromWIF(wifString) {
        const wasm = WasmECPair.from_wif(wifString);
        return new ECPair(wasm);
    }
    /**
     * Create an ECPair from a mainnet WIF string
     * @param wifString - The WIF-encoded private key string
     * @returns An ECPair instance
     */
    static fromWIFMainnet(wifString) {
        const wasm = WasmECPair.from_wif_mainnet(wifString);
        return new ECPair(wasm);
    }
    /**
     * Create an ECPair from a testnet WIF string
     * @param wifString - The WIF-encoded private key string
     * @returns An ECPair instance
     */
    static fromWIFTestnet(wifString) {
        const wasm = WasmECPair.from_wif_testnet(wifString);
        return new ECPair(wasm);
    }
    /**
     * Get the private key as a Uint8Array (if available)
     */
    get privateKey() {
        return this._wasm.private_key;
    }
    /**
     * Get the public key as a Uint8Array
     */
    get publicKey() {
        return this._wasm.public_key;
    }
    /**
     * Convert to WIF string (mainnet)
     * @returns The WIF-encoded private key
     */
    toWIF() {
        return this._wasm.to_wif();
    }
    /**
     * Convert to mainnet WIF string
     * @returns The WIF-encoded private key
     */
    toWIFMainnet() {
        return this._wasm.to_wif_mainnet();
    }
    /**
     * Convert to testnet WIF string
     * @returns The WIF-encoded private key
     */
    toWIFTestnet() {
        return this._wasm.to_wif_testnet();
    }
    /**
     * Get the underlying WASM instance (internal use only)
     * @internal
     */
    get wasm() {
        return this._wasm;
    }
}
