// @bitgo/sdk-coin-soneium/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCommon = getCommon;
exports.recoveryBlockchainExplorerQuery = recoveryBlockchainExplorerQuery;
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const resources_1 = require("./resources");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const superagent_1 = __importDefault(require("superagent"));
const ethereumjs_util_1 = require("ethereumjs-util");
const RESPONSE_STATUS_ERROR = '0';
const RESPONSE_MESSAGE_NOTOK = 'NOTOK';
const commons = new Map([
    [statics_1.NetworkType.MAINNET, resources_1.mainnetCommon],
    [statics_1.NetworkType.TESTNET, resources_1.testnetCommon],
]);
/**
 * @param {NetworkType} network either mainnet or testnet
 * @returns {EthereumCommon} Ethereum common configuration object
 */
function getCommon(network) {
    const common = commons.get(network);
    if (!common) {
        throw new sdk_core_1.InvalidTransactionError('Missing network common configuration');
    }
    return common;
}
/**
 * Handle standard blockchain explorer query
 * @param query Query parameters
 * @param bitgoEnv BitGo environment
 * @returns Response from soneium.network
 */
async function handleStandardExplorerQuery(query, bitgoEnv) {
    const apiToken = sdk_core_1.common.Environments[bitgoEnv].soneiumExplorerApiToken;
    const explorerUrl = sdk_core_1.common.Environments[bitgoEnv].soneiumExplorerBaseUrl;
    return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
}
/**
 * Handle proxy blockchain explorer query
 * @param query Query parameters
 * @param bitgoEnv BitGo environment
 * @returns Response from soneium.network
 */
async function handleProxyExplorerQuery(query, bitgoEnv) {
    const body = {
        jsonrpc: '2.0',
        method: query.action,
        params: [
            {
                to: query.to,
                data: (0, ethereumjs_util_1.addHexPrefix)(query.data),
            },
            query.tag || 'latest',
        ],
        id: 1,
    };
    const response = await superagent_1.default.post(sdk_core_1.common.Environments[bitgoEnv].soneiumExplorerBaseUrl + '/api/eth-rpc').send(body);
    if (!response.ok) {
        throw new Error('Could not reach soneium.network');
    }
    if (response.body.status === RESPONSE_STATUS_ERROR && response.body.message === RESPONSE_MESSAGE_NOTOK) {
        throw new Error('Soneium.network rate limit reached');
    }
    return response.body;
}
/**
 * Make a query to soneium.network for information such as balance, token balance, solidity calls
 * @param query Query parameters
 * @param bitgoEnv BitGo environment
 * @returns Response from soneium.network
 */
async function recoveryBlockchainExplorerQuery(query, bitgoEnv) {
    try {
        if (query.module === 'proxy') {
            return await handleProxyExplorerQuery(query, bitgoEnv);
        }
        return await handleStandardExplorerQuery(query, bitgoEnv);
    }
    catch (error) {
        throw new Error(`Could not query soneium explorer, error: ${error?.message || 'Unknown error'}`);
    }
}
