// @bitgo/public-types/src/schema/tss/baseTypes.ts
export declare enum SignatureShareType {
  USER = "user",
  BACKUP = "backup",
  BITGO = "bitgo",
}

export declare enum EncryptedSignerShareType {
  ENCRYPTED_SIGNER_SHARE = "encryptedSignerShare",
  ENCRYPTED_R_SHARE = "encryptedRShare",
}

export declare enum CommitmentType {
  COMMITMENT = "commitment",
  DECOMMITMENT = "decommitment",
}

interface ShareBaseRecord {
  from: SignatureShareType;
  to: SignatureShareType;
  share: string;
}

export interface SignatureShareRecord extends ShareBaseRecord {
  vssProof?: string;
  privateShareProof?: string;
  publicShare?: string;
}

export interface EncryptedSignerShareRecord extends ShareBaseRecord {
  type: EncryptedSignerShareType;
}

export interface CommitmentShareRecord extends ShareBaseRecord {
  type: CommitmentType;
}

export interface Signature {
  y: string;
  R: string;
  sigma: string;
}

export type BaseWalletType =
  | "backing"
  | "cold"
  | "custodial"
  | "custodialPaired"
  | "hot"
  | "trading"
  | "advanced";

export type TxRequestState =
  | "pendingCommitment"
  | "pendingApproval"
  | "canceled"
  | "rejected"
  | "initialized"
  | "pendingDelivery"
  | "delivered"
  | "pendingUserSignature"
  | "signed";

export type UnsignedMessageTss = {
  derivationPath: string;
  message: string;
};

export type SignableTransaction = {
  serializedTxHex: string;
  signableHex: string;
};

export type UnsignedTransactionTss = SignableTransaction & {
  derivationPath: string;
  feeInfo?: {
    fee: number;
    feeString: string;
  };
  coinSpecific?: Record<string, unknown>;
  parsedTx?: unknown;
};

export type BaseTransactionState =
  | "initialized"
  | "pendingCommitment"
  | "pendingSignature"
  | "signed"
  | "held"
  | "delivered"
  | "invalidSignature"
  | "rejected";

export type SignedTx = {
  id: string;
  tx: string;
  publicKey?: string;
  signature?: string;
};

export type TxRequestVersion = "full" | "lite";

export type TxRequest = {
  txRequestId: string;
  walletId: string;
  walletType: BaseWalletType;
  version: number;
  enterpriseId?: string;
  state: TxRequestState;
  date: string;
  userId: string;
  intent: unknown;
  pendingApprovalId?: string;
  policiesChecked: boolean;
  signatureShares?: SignatureShareRecord[];
  commitmentShares?: CommitmentShareRecord[];
  pendingTxHashes?: string[];
  txHashes?: string[];
  unsignedMessages?: UnsignedMessageTss[];
  unsignedTxs: UnsignedTransactionTss[];
  transactions?: {
    state: BaseTransactionState;
    unsignedTx: UnsignedTransactionTss;
    signatureShares: SignatureShareRecord[];
    signedTx?: SignedTx;
    commitmentShares?: CommitmentShareRecord[];
  }[];
  messages?: {
    state: BaseTransactionState;
    signatureShares: SignatureShareRecord[];
    messageRaw: string;
    messageEncoded?: string;
    derivationPath: string;
    combineSigShare?: string;
    txHash?: string;
  }[];
  apiVersion?: TxRequestVersion;
  latest: boolean;
};
