// @bitgo/abstract-lightning/dist/src/abstractLightningCoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractLightningCoin = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const crypto_1 = require("crypto");
const utxo_lib_1 = require("@bitgo/utxo-lib");
class AbstractLightningCoin extends sdk_core_1.BaseCoin {
    constructor(bitgo, network) {
        super(bitgo);
        this._network = network;
    }
    getBaseFactor() {
        return 10e11;
    }
    verifyTransaction(params) {
        throw new Error('Method not implemented.');
    }
    isWalletAddress(params) {
        throw new Error('Method not implemented.');
    }
    parseTransaction(params) {
        throw new Error('Method not implemented.');
    }
    generateKeyPair(seed) {
        if (!seed) {
            // An extended private key has both a normal 256 bit private key and a 256
            // bit chain code, both of which must be random. 512 bits is therefore the
            // maximum entropy and gives us maximum security against cracking.
            seed = (0, crypto_1.randomBytes)(512 / 8);
        }
        const extendedKey = utxo_lib_1.bip32.fromSeed(seed);
        return {
            pub: extendedKey.neutered().toBase58(),
            prv: extendedKey.toBase58(),
        };
    }
    isValidPub(pub) {
        throw new Error('Method not implemented.');
    }
    isValidAddress(address) {
        throw new Error('Method not implemented.');
    }
    signTransaction(params) {
        throw new Error('Method not implemented.');
    }
    /** @inheritDoc */
    auditDecryptedKey(params) {
        /** https://bitgoinc.atlassian.net/browse/BTC-2149 */
        throw new Error('Method not implemented.');
    }
}
exports.AbstractLightningCoin = AbstractLightningCoin;
