// @bitgo/abstract-utxo/dist/cjs/test/unit/offlineVault/halfSigned.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const assert_1 = __importDefault(require("assert"));
const crypto_1 = __importDefault(require("crypto"));
const t = __importStar(require("io-ts"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const offlineVault_1 = require("../../../src/offlineVault");
const descriptor_1 = require("../../../src/offlineVault/descriptor");
const fixtures_utils_1 = require("../transaction/descriptor/fixtures.utils");
const { assertEqualFixture } = (0, fixtures_utils_1.getFixtureRoot)(__dirname + '/fixtures');
function getFixturesNames() {
    // I'm using sync here because mocha cannot do async setup
    // eslint-disable-next-line no-sync
    return fs.readdirSync(__dirname + '/fixtures').filter((f) => f.endsWith('.json') && !f.endsWith('.explanation.json'));
}
const KeyPair = t.intersection([t.type({ xpub: t.string }), t.partial({ xprv: t.string })]);
const KeyWithParent = t.intersection([KeyPair, t.partial({ parent: KeyPair })]);
const Fixture = t.type({
    walletKeys: t.array(KeyWithParent),
    response: t.unknown,
});
async function readFixture(name) {
    const data = JSON.parse(await fs.promises.readFile(__dirname + '/fixtures/' + name, 'utf-8'));
    return (0, sdk_core_1.decodeOrElse)('Fixture', Fixture, data, (e) => {
        throw new Error(`failed to decode fixture ${name}: ${e}`);
    });
}
function withRotatedXpubs(tx) {
    const { user, backup, bitgo } = tx.xpubsWithDerivationPath;
    return {
        ...tx,
        xpubsWithDerivationPath: {
            user: bitgo,
            backup: user,
            bitgo: backup,
        },
    };
}
function withRandomXpubs(tx) {
    function randomXpub() {
        const bytes = crypto_1.default.getRandomValues(new Uint8Array(32));
        return utxolib.bip32.fromSeed(Buffer.from(bytes)).neutered().toBase58();
    }
    return {
        ...tx,
        xpubsWithDerivationPath: {
            user: randomXpub(),
            backup: randomXpub(),
            bitgo: randomXpub(),
        },
    };
}
function withoutDescriptors(tx) {
    return {
        ...tx,
        descriptors: [],
    };
}
function getDerivationId(v) {
    const id = v.user.derivedFromParentWithSeed;
    (0, assert_1.default)(id);
    return id;
}
function getRootPrv(walletKeys) {
    (0, assert_1.default)(walletKeys[0]);
    (0, assert_1.default)(walletKeys[0].parent);
    (0, assert_1.default)(walletKeys[0].parent.xprv);
    return utxolib.bip32.fromBase58(walletKeys[0].parent.xprv);
}
describe('OfflineVaultHalfSigned', function () {
    for (const fixtureName of getFixturesNames()) {
        it(`can sign fixture ${fixtureName}`, async function () {
            const { walletKeys, response } = await readFixture(fixtureName);
            (0, assert_1.default)(descriptor_1.DescriptorTransaction.is(response));
            const rootPrv = getRootPrv(walletKeys);
            const derivationId = getDerivationId(response.xpubsWithDerivationPath);
            (0, offlineVault_1.createHalfSigned)('btc', rootPrv, derivationId, response);
            const mutations = [withRotatedXpubs(response), withRandomXpubs(response), withoutDescriptors(response)];
            for (const mutation of mutations) {
                assert_1.default.throws(() => (0, offlineVault_1.createHalfSigned)('btc', rootPrv, derivationId, mutation));
            }
            await assertEqualFixture(fixtureName.replace(/\.json$/, '.explanation.json'), (0, offlineVault_1.getTransactionExplanation)('btc', response));
        });
    }
});
