// @bitgo/public-types/src/schema/transactions/coinSpecific.ts
import * as t from "io-ts";
import {
  optionalStringOrNumber,
  optionalString,
  optionalNumber,
  optionalStringArray,
  optionalBoolean,
  optionalUnknown,
} from "../../utils";

const BaseBuildParams = {
  type: t.string,
  nonce: optionalNumber,
  memo: t.partial({
    type: optionalString,
    value: optionalString,
  }),
  comment: optionalString,
  ledgerSequenceDelta: optionalNumber,
  lastLedgerSequence: optionalNumber,
  feeRate: optionalStringOrNumber,
  messageKey: optionalString,
  previewPendingTxs: optionalBoolean,
  pendingApprovalId: optionalString,
  sequenceId: optionalString,
};

const ConsolidateBuildParams = {
  ...BaseBuildParams,
  consolidateId: optionalString,
  senderAddress: optionalString,
};

// Coin Specific Build Params
export const CoinSpecificBuildParams = {
  utxo: {
    /** Transaction Locktime */
    locktime: optionalNumber,
    limit: optionalNumber,
    bulk: optionalBoolean,
  },
  avaxp: {
    locktime: optionalNumber,
    sourceChain: optionalString,
    destinationChain: optionalString,
    unspents: optionalStringArray,
  },
  dot: {
    ...ConsolidateBuildParams,
    sequenceId: optionalString,
    senderAddressIndex: optionalNumber,
    nonce: optionalNumber,
    proxyType: optionalString,
    forceProxyType: optionalString,
    owner: optionalUnknown,
    delay: optionalUnknown,
    index: optionalUnknown,
    validityWindow: optionalUnknown,
    version: optionalUnknown,
    referenceBlock: optionalUnknown,
  },
  sol: {
    tokenName: optionalString,
  },
};
