// @bitgo/public-types/src/schema/transactions/tronTransaction.ts
import * as t from "io-ts";
import { optionalNumber } from "../../utils";

export const TronResourceType = t.keyof({
  BANDWIDTH: 1,
  ENERGY: 1,
});
export type TronResourceType = t.TypeOf<typeof TronResourceType>;

export const TronFreezeTransaction = t.type({
  type: t.literal("freeze"),
  owner_address: t.string,
  frozen_balance: t.string,
  resource: TronResourceType,
});

export const TronUnfreezeTransaction = t.type({
  type: t.literal("unfreeze"),
  owner_address: t.string,
  unfreeze_balance: t.string,
  resource: TronResourceType,
});

export const TronVoteTransaction = t.type({
  type: t.literal("vote"),
  owner_address: t.string,
  votes: t.array(
    t.type({
      vote_address: t.string,
      vote_count: t.string,
    }),
  ),
});

export const TronDelegateResourceTransaction = t.intersection([
  t.type({
    type: t.literal("delegateResource"),
    owner_address: t.string,
    resource: TronResourceType,
    receiver_address: t.string,
    balance: t.string,
  }),
  t.partial({
    lock: t.boolean,
    lock_period: optionalNumber,
  }),
]);

export const TronUndelegateResourceTransaction = t.type({
  type: t.literal("undelegateResource"),
  owner_address: t.string,
  resource: TronResourceType,
  receiver_address: t.string,
  balance: t.string,
});

export const TronClaimRewardsTransaction = t.type({
  type: t.literal("claimRewards"),
  owner_address: t.string,
});

export const TronWithdrawExpireUnfreezeTransaction = t.type({
  type: t.literal("withdrawExpireUnfreeze"),
  owner_address: t.string,
});

export const TronTransaction = t.union([
  TronFreezeTransaction,
  TronUnfreezeTransaction,
  TronVoteTransaction,
  TronDelegateResourceTransaction,
  TronUndelegateResourceTransaction,
  TronClaimRewardsTransaction,
  TronWithdrawExpireUnfreezeTransaction,
]);

export type TronFreezeTransaction = t.TypeOf<typeof TronFreezeTransaction>;
export type TronUnfreezeTransaction = t.TypeOf<typeof TronUnfreezeTransaction>;
export type TronVoteTransaction = t.TypeOf<typeof TronVoteTransaction>;
export type TronDelegateResourceTransaction = t.TypeOf<
  typeof TronDelegateResourceTransaction
>;
export type TronUndelegateResourceTransaction = t.TypeOf<
  typeof TronUndelegateResourceTransaction
>;
export type TronClaimRewardsTransaction = t.TypeOf<
  typeof TronClaimRewardsTransaction
>;
export type TronWithdrawExpireUnfreezeTransaction = t.TypeOf<
  typeof TronWithdrawExpireUnfreezeTransaction
>;
export type TronTransaction = t.TypeOf<typeof TronTransaction>;
