// @bitgo/utxo-ord/dist/cjs/src/psbt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorNoLayout = exports.MAX_UNSPENTS_FOR_OUTPUT_LAYOUT = exports.DefaultInscriptionConstraints = void 0;
exports.createPsbtFromOutputLayout = createPsbtFromOutputLayout;
exports.findOutputLayoutForWalletUnspents = findOutputLayoutForWalletUnspents;
exports.createPsbtForSingleInscriptionPassingTransaction = createPsbtForSingleInscriptionPassingTransaction;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const OrdOutput_1 = require("./OrdOutput");
const SatPoint_1 = require("./SatPoint");
const SatRange_1 = require("./SatRange");
const OutputLayout_1 = require("./OutputLayout");
const combinations_1 = require("./combinations");
const { BitGoPsbt, ChainCode } = wasm_utxo_1.fixedScriptWallet;
/**
 * Map utxo-lib network objects to CoinName strings.
 */
function networkToCoinName(network) {
    // Bitcoin mainnet
    if (network.bech32 === 'bc' && network.pubKeyHash === 0) {
        return 'btc';
    }
    // Bitcoin testnet
    if (network.bech32 === 'tb' && network.pubKeyHash === 111) {
        return 'tbtc';
    }
    throw new Error(`Unknown network: ${JSON.stringify(network)}`);
}
/**
 * Normalize network parameter - accepts either CoinName string or utxo-lib Network object.
 */
function normalizeCoinName(networkOrCoinName) {
    if (typeof networkOrCoinName === 'string') {
        return networkOrCoinName;
    }
    return networkToCoinName(networkOrCoinName);
}
/** Segwit transaction overhead in virtual bytes */
const TX_SEGWIT_OVERHEAD_VSIZE = 10;
exports.DefaultInscriptionConstraints = {
    minChangeOutput: BigInt(10000),
    minInscriptionOutput: BigInt(10000),
    maxInscriptionOutput: BigInt(20000),
};
function createPsbtFromOutputLayout(networkOrCoinName, inputBuilder, unspents, outputs, outputLayout) {
    if (unspents.length === 0) {
        throw new Error(`must provide at least one unspent`);
    }
    const coinName = normalizeCoinName(networkOrCoinName);
    const psbt = BitGoPsbt.createEmpty(coinName, inputBuilder.walletKeys);
    // Add inputs
    unspents.forEach((u) => {
        const { txid, vout } = (0, SatPoint_1.parseOutputId)(u.id);
        psbt.addWalletInput({ txid, vout, value: u.value }, inputBuilder.walletKeys, {
            scriptId: { chain: u.chain, index: u.index },
            signPath: { signer: inputBuilder.signer, cosigner: inputBuilder.cosigner },
        });
    });
    // Build ord outputs from layout
    const ordInput = OrdOutput_1.OrdOutput.joinAll(unspents.map((u) => new OrdOutput_1.OrdOutput(u.value)));
    const ordOutputs = (0, OutputLayout_1.getOrdOutputsForLayout)(ordInput, outputLayout);
    (0, OutputLayout_1.toArray)(ordOutputs).forEach((ordOutput) => {
        if (ordOutput === null) {
            return;
        }
        switch (ordOutput) {
            // skip fee output (virtual)
            case ordOutputs.feeOutput:
                return;
            // add padding/change outputs
            case ordOutputs.firstChangeOutput:
            case ordOutputs.secondChangeOutput:
                const { chain, index } = ordOutput === ordOutputs.firstChangeOutput ? outputs.changeOutputs[0] : outputs.changeOutputs[1];
                psbt.addWalletOutput(inputBuilder.walletKeys, { chain, index, value: ordOutput.value });
                break;
            // add actual inscription output
            case ordOutputs.inscriptionOutput:
                const recipient = outputs.inscriptionRecipient;
                if (typeof recipient === 'string') {
                    psbt.addOutput(recipient, ordOutput.value);
                }
                else if (recipient instanceof Uint8Array) {
                    psbt.addOutput(recipient, ordOutput.value);
                }
                else {
                    throw new Error('inscriptionRecipient must be a string or Uint8Array');
                }
                break;
        }
    });
    return psbt;
}
function toSatRange(p) {
    const { offset } = (0, SatPoint_1.parseSatPoint)(p);
    return new SatRange_1.SatRange(offset, offset);
}
function getFee(vsize, rateSatPerKB) {
    return BigInt(Math.ceil((vsize * rateSatPerKB) / 1000));
}
/**
 * @param inputs - inscription input must come first
 * @param satPoint - location of the inscription
 * @param outputs
 * @param constraints
 * @param minimizeInputs
 */
function findOutputLayoutForWalletUnspents(inputs, satPoint, outputs, constraints, { minimizeInputs = false } = {}) {
    if (minimizeInputs) {
        return findSmallestOutputLayoutForWalletUnspents(inputs, satPoint, outputs, constraints);
    }
    if (inputs.length === 0) {
        throw new Error(`must provide at least one input`);
    }
    if (outputs.changeOutputs[0].chain !== outputs.changeOutputs[1].chain) {
        // otherwise our fee calc is too complicated
        throw new Error(`wallet outputs must be on same chain`);
    }
    const { minChangeOutput = exports.DefaultInscriptionConstraints.minChangeOutput, minInscriptionOutput = exports.DefaultInscriptionConstraints.minInscriptionOutput, maxInscriptionOutput = exports.DefaultInscriptionConstraints.maxInscriptionOutput, } = constraints;
    // Calculate input vsize using wasm-utxo Dimensions
    const inputDimensions = inputs.reduce((dims, input) => dims.plus(wasm_utxo_1.Dimensions.fromInput({ chain: input.chain })), wasm_utxo_1.Dimensions.empty());
    const inputsVSize = inputDimensions.getInputVSize();
    // Calculate output vsize using wasm-utxo Dimensions
    const outputDimensions = wasm_utxo_1.Dimensions.fromOutput({
        scriptType: ChainCode.scriptType(outputs.changeOutputs[0].chain),
    });
    const outputVSize = outputDimensions.getOutputVSize();
    // Join all the inputs into a single inscriptionOutput.
    // For the purposes of finding a layout there is no difference.
    const inscriptionOutput = OrdOutput_1.OrdOutput.joinAll(inputs.map((i) => new OrdOutput_1.OrdOutput(i.value, i === inputs[0] ? [toSatRange(satPoint)] : [])));
    const layout = (0, OutputLayout_1.findOutputLayout)(inscriptionOutput, {
        minChangeOutput,
        minInscriptionOutput,
        maxInscriptionOutput,
        feeFixed: getFee(TX_SEGWIT_OVERHEAD_VSIZE + inputsVSize, constraints.feeRateSatKB),
        feePerOutput: getFee(outputVSize, constraints.feeRateSatKB),
    });
    return layout ? { inputs, layout } : undefined;
}
exports.MAX_UNSPENTS_FOR_OUTPUT_LAYOUT = 5;
/**
 * @param inputs - inscription input must come first
 * @param satPoint - location of the inscription
 * @param outputs
 * @param constraints
 */
function findSmallestOutputLayoutForWalletUnspents(inputs, satPoint, outputs, constraints) {
    if (exports.MAX_UNSPENTS_FOR_OUTPUT_LAYOUT < inputs.length) {
        throw new Error(`input array is too large`);
    }
    // create powerset of all supplementary inputs and find the cheapest result
    const inputsArr = [inputs, ...(0, combinations_1.powerset)(inputs.slice(1)).map((s) => [inputs[0], ...s])];
    return inputsArr
        .map((inputs) => findOutputLayoutForWalletUnspents(inputs, satPoint, outputs, constraints))
        .reduce((best, next) => {
        if (best === undefined) {
            return next;
        }
        if (next === undefined) {
            return best;
        }
        return best.layout.feeOutput < next.layout.feeOutput ? best : next;
    });
}
class ErrorNoLayout extends Error {
    constructor() {
        super('Could not find output layout for inscription passing transaction');
    }
}
exports.ErrorNoLayout = ErrorNoLayout;
/**
 * @param networkOrCoinName - Coin name (e.g., "btc", "tbtc") or utxo-lib Network object
 * @param inputBuilder
 * @param unspent
 * @param satPoint
 * @param outputs
 * @param constraints
 * @param supplementaryUnspents - additional inputs to cover fee.
 * @param [minimizeInputs=true] - try to find input combination with minimal fees. Limits supplementaryUnspents to 4.
 */
function createPsbtForSingleInscriptionPassingTransaction(networkOrCoinName, inputBuilder, unspent, satPoint, outputs, constraints, { supplementaryUnspents = [], minimizeInputs = true, } = {}) {
    // support for legacy call style
    if (Array.isArray(unspent)) {
        if (unspent.length !== 1) {
            throw new Error(`can only pass single unspent`);
        }
        unspent = unspent[0];
    }
    const result = findOutputLayoutForWalletUnspents([unspent, ...supplementaryUnspents], satPoint, outputs, constraints, { minimizeInputs });
    if (!result) {
        throw new ErrorNoLayout();
    }
    return createPsbtFromOutputLayout(networkOrCoinName, inputBuilder, result.inputs, outputs, result.layout);
}
