// @bitgo/utxo-ord/dist/cjs/src/SatRange.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SatRange = exports.InvalidSatRange = void 0;
class InvalidSatRange extends Error {
    constructor(message, start, end) {
        super(message);
        this.start = start;
        this.end = end;
    }
}
exports.InvalidSatRange = InvalidSatRange;
function toRange(v) {
    return typeof v === 'bigint' ? new SatRange(v, v) : v;
}
/**
 * Range of satoshi, inclusive.
 * Inscriptions have start === end.
 */
class SatRange {
    constructor(start, end) {
        this.start = start;
        this.end = end;
        if (start < 0 || end < 0 || end < start) {
            throw new InvalidSatRange(`Invalid SatRange [${start}, ${end}]`, start, end);
        }
    }
    /**
     * @param offset
     * @return SatRange with start and end shifted by offset
     */
    shiftedBy(offset) {
        return new SatRange(this.start + offset, this.end + offset);
    }
    /** @return true iff this intersects with _other_ */
    intersectsWith(other) {
        other = toRange(other);
        if (this.start <= other.start) {
            return other.start <= this.end;
        }
        return other.intersectsWith(this);
    }
    /** @return true iff this is superset of _other_. */
    isSupersetOf(other) {
        other = toRange(other);
        return this.start <= other.start && other.end <= this.end;
    }
    toString() {
        if (this.start === this.end) {
            return `[${this.start}]`;
        }
        return `[${this.start}..${this.end}]`;
    }
    size() {
        return this.end - this.start + BigInt(1);
    }
}
exports.SatRange = SatRange;
