// @bitgo/statics/dist/src/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tip20Token = exports.CantonToken = exports.AdaToken = exports.PolyxCoin = exports.TaoCoin = exports.JettonToken = exports.CosmosChainToken = exports.VetToken = exports.Nep141Token = exports.Sip10Token = exports.FiatCoin = exports.VetNFTCollection = exports.AptNFTCollection = exports.AptCoin = exports.SuiCoin = exports.XrpCoin = exports.MonERC20Token = exports.XdcERC20Token = exports.FlrERC20Token = exports.WorldERC20Token = exports.CoredaoERC20Token = exports.BeraERC20Token = exports.ZkethERC20Token = exports.OpethERC20Token = exports.ArbethERC20Token = exports.PolygonERC20Token = exports.AvaxERC20Token = exports.EthLikeERC721Token = exports.EthLikeERC20Token = exports.SolCoin = exports.EosCoin = exports.AlgoCoin = exports.HederaToken = exports.HederaCoin = exports.StellarCoin = exports.BscCoin = exports.CeloCoin = exports.Erc20CompatibleAccountCoin = exports.TronErc20Coin = exports.Erc1155Coin = exports.Erc721Coin = exports.Erc20Coin = exports.Base58ContractAddressDefinedToken = exports.NFTCollectionIdDefinedToken = exports.ContractAddressDefinedToken = exports.GasTankAccountCoin = exports.AccountCoinToken = exports.AccountCoin = exports.ProgramID = void 0;
exports.account = account;
exports.gasTankAccount = gasTankAccount;
exports.erc20Token = erc20Token;
exports.erc721Token = erc721Token;
exports.erc20 = erc20;
exports.terc20 = terc20;
exports.erc721 = erc721;
exports.terc721 = terc721;
exports.nonstandardToken = nonstandardToken;
exports.erc1155 = erc1155;
exports.terc1155 = terc1155;
exports.erc20CompatibleAccountCoin = erc20CompatibleAccountCoin;
exports.celoToken = celoToken;
exports.tceloToken = tceloToken;
exports.bscToken = bscToken;
exports.tbscToken = tbscToken;
exports.stellarToken = stellarToken;
exports.tstellarToken = tstellarToken;
exports.tronToken = tronToken;
exports.ttronToken = ttronToken;
exports.hederaCoin = hederaCoin;
exports.hederaToken = hederaToken;
exports.algoToken = algoToken;
exports.talgoToken = talgoToken;
exports.eosToken = eosToken;
exports.teosToken = teosToken;
exports.solToken = solToken;
exports.tsolToken = tsolToken;
exports.adaToken = adaToken;
exports.tadaToken = tadaToken;
exports.avaxErc20 = avaxErc20;
exports.tavaxErc20 = tavaxErc20;
exports.polygonErc20 = polygonErc20;
exports.tpolygonErc20 = tpolygonErc20;
exports.arbethErc20 = arbethErc20;
exports.tarbethErc20 = tarbethErc20;
exports.opethErc20 = opethErc20;
exports.topethErc20 = topethErc20;
exports.zkethErc20 = zkethErc20;
exports.tzkethErc20 = tzkethErc20;
exports.beraErc20 = beraErc20;
exports.tberaErc20 = tberaErc20;
exports.coredaoErc20 = coredaoErc20;
exports.tcoredaoErc20 = tcoredaoErc20;
exports.worldErc20 = worldErc20;
exports.tworldErc20 = tworldErc20;
exports.flrErc20 = flrErc20;
exports.tflrErc20 = tflrErc20;
exports.xdcErc20 = xdcErc20;
exports.txdcErc20 = txdcErc20;
exports.monErc20 = monErc20;
exports.tmonErc20 = tmonErc20;
exports.xrpToken = xrpToken;
exports.txrpToken = txrpToken;
exports.suiToken = suiToken;
exports.tsuiToken = tsuiToken;
exports.aptToken = aptToken;
exports.aptNFTCollection = aptNFTCollection;
exports.taptToken = taptToken;
exports.taptNFTCollection = taptNFTCollection;
exports.fiat = fiat;
exports.sip10Token = sip10Token;
exports.tsip10Token = tsip10Token;
exports.nep141Token = nep141Token;
exports.tnep141Token = tnep141Token;
exports.vetToken = vetToken;
exports.tvetToken = tvetToken;
exports.vetNFTCollection = vetNFTCollection;
exports.tvetNFTCollection = tvetNFTCollection;
exports.cosmosToken = cosmosToken;
exports.taoToken = taoToken;
exports.ttaoToken = ttaoToken;
exports.polyxToken = polyxToken;
exports.tpolyxToken = tpolyxToken;
exports.jettonToken = jettonToken;
exports.tjettonToken = tjettonToken;
exports.cantonToken = cantonToken;
exports.tcantonToken = tcantonToken;
exports.tip20Token = tip20Token;
exports.ttip20Token = ttip20Token;
const base_1 = require("./base");
const constants_1 = require("./constants");
const errors_1 = require("./errors");
const networks_1 = require("./networks");
const coinFeatures_1 = require("./coinFeatures");
/**
 * This is the program id against sol token program.
 */
var ProgramID;
(function (ProgramID) {
    ProgramID["TokenProgramId"] = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";
    ProgramID["Token2022ProgramId"] = "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb";
})(ProgramID || (exports.ProgramID = ProgramID = {}));
/**
 * Account based coins, such as Ethereum, Stellar, or XRP.
 *
 * These types of coins maintain an "account balance" for each address on the network,
 * as opposed to the unspent transaction output model which maintains a record of all
 * "pieces" of coin which belong to an address.
 */
class AccountCoin extends base_1.BaseCoin {
    constructor(options) {
        super({
            ...options,
            kind: base_1.CoinKind.CRYPTO,
        });
        this.network = options.network;
    }
    requiredFeatures() {
        return new Set([base_1.CoinFeature.ACCOUNT_MODEL]);
    }
    disallowedFeatures() {
        return new Set([base_1.CoinFeature.UNSPENT_MODEL]);
    }
    static getFeaturesExcluding(excludedFeatures) {
        return AccountCoin.DEFAULT_FEATURES.filter((feature) => !excludedFeatures.includes(feature));
    }
    static getFeaturesByTypeExcluding(excludedFeatures, baseFeatures = AccountCoin.DEFAULT_FEATURES) {
        return baseFeatures.filter((feature) => !excludedFeatures.includes(feature));
    }
}
exports.AccountCoin = AccountCoin;
AccountCoin.DEFAULT_FEATURES = coinFeatures_1.ACCOUNT_COIN_DEFAULT_FEATURES;
// Need to gate some high risk coin from SINGAPORE trust
AccountCoin.DEFAULT_FEATURES_EXCLUDE_SINGAPORE = coinFeatures_1.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE;
class AccountCoinToken extends AccountCoin {
    constructor(options) {
        super({
            ...options,
        });
    }
}
exports.AccountCoinToken = AccountCoinToken;
class GasTankAccountCoin extends AccountCoin {
    constructor(options) {
        super({
            ...options,
        });
        this.gasTankLowBalanceAlertFactor = options.gasTankLowBalanceAlertFactor;
        this.gasTankMinBalanceRecommendationFactor = options.gasTankMinBalanceRecommendationFactor;
        this.gasTankToken = options?.gasTankToken;
    }
}
exports.GasTankAccountCoin = GasTankAccountCoin;
/**
 * Some blockchains support tokens which are defined by an address at which they have a smart contract deployed.
 * Examples are ERC20 tokens, and the equivalent on other chains.
 */
class ContractAddressDefinedToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        // valid ERC 20 contract addresses are "0x" followed by 40 lowercase hex characters
        // do not use a valid address format for generic tokens because they not have onchain addresses
        if (!options.contractAddress.match(/^0x[a-f0-9]{40}$/) && !options.features.includes(base_1.CoinFeature.GENERIC_TOKEN)) {
            throw new errors_1.InvalidContractAddressError(options.name, options.contractAddress);
        }
        this.contractAddress = options.contractAddress;
    }
}
exports.ContractAddressDefinedToken = ContractAddressDefinedToken;
/**
 * Used for blockchains that support NFT collections.
 */
class NFTCollectionIdDefinedToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.nftCollectionId = options.nftCollectionId;
    }
}
exports.NFTCollectionIdDefinedToken = NFTCollectionIdDefinedToken;
/**
 * ERC20 token addresses are Base58 formatted on some blockchains.
 */
class Base58ContractAddressDefinedToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        if (!/^[1-9A-HJ-NP-Za-km-z]{34}$/.test(options.contractAddress)) {
            throw new errors_1.InvalidContractAddressError(options.name, options.contractAddress);
        }
        this.contractAddress = options.contractAddress;
    }
}
exports.Base58ContractAddressDefinedToken = Base58ContractAddressDefinedToken;
/**
 * ERC 20 is a token standard for the Ethereum blockchain. They are similar to other account coins, but have a
 * contract address property which identifies the smart contract which defines the token.
 */
class Erc20Coin extends ContractAddressDefinedToken {
}
exports.Erc20Coin = Erc20Coin;
/**
 * ERC 721 is the non fungible token standard for the Ethereum blockchain.
 *
 * {@link https://eips.ethereum.org/EIPS/eip-721 EIP721}
 */
class Erc721Coin extends ContractAddressDefinedToken {
}
exports.Erc721Coin = Erc721Coin;
/**
 * ERC 1155 is the multi token standard for the Ethereum blockchain.
 *
 * {@link https://eips.ethereum.org/EIPS/eip-1155 EIP1155}
 */
class Erc1155Coin extends ContractAddressDefinedToken {
}
exports.Erc1155Coin = Erc1155Coin;
/**
 * The TRON blockchain supports tokens of the ERC20 standard similar to ETH ERC20 tokens.
 */
class TronErc20Coin extends Base58ContractAddressDefinedToken {
}
exports.TronErc20Coin = TronErc20Coin;
/**
 * Some blockchains have native coins which also support the ERC20 interface such as CELO.
 */
class Erc20CompatibleAccountCoin extends ContractAddressDefinedToken {
    constructor(options) {
        super({
            ...options,
            // These coins should not be classified as tokens as they are not children of other coins
            isToken: false,
        });
    }
}
exports.Erc20CompatibleAccountCoin = Erc20CompatibleAccountCoin;
/**
 * The CELO blockchain supports tokens of the ERC20 standard similar to ETH ERC20 tokens.
 */
class CeloCoin extends ContractAddressDefinedToken {
}
exports.CeloCoin = CeloCoin;
/**
 * The BSC blockchain supports tokens of the ERC20 standard similar to ETH ERC20 tokens.
 */
class BscCoin extends ContractAddressDefinedToken {
}
exports.BscCoin = BscCoin;
/**
 * The Stellar network supports tokens (non-native assets)
 * XLM is also known as the native asset.
 * Stellar tokens work similar to XLM, but the token name is determined by the chain,
 * the token code and the issuer account in the form: (t)xlm:<token>-<issuer>
 */
class StellarCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        if (options.domain !== '' && !options.domain.match(constants_1.DOMAIN_PATTERN)) {
            throw new errors_1.InvalidDomainError(options.name, options.domain);
        }
        this.domain = options.domain;
    }
}
exports.StellarCoin = StellarCoin;
/**
 * The Hedera coin needs a client set with the node account Id.
 * It's an account based coin that needs the node account ID
 * where the transaction will be sent.
 *
 */
class HederaCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.nodeAccountId = options.nodeAccountId;
    }
}
exports.HederaCoin = HederaCoin;
/**
 * The Hedera network supports tokens.
 * Hedera tokens work similar to native Hedera coin,
 * but the token is determined by the tokenId on the node
 *
 */
class HederaToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.nodeAccountId = options.nodeAccountId;
        this.tokenId = options.contractAddress;
        this.contractAddress = options.contractAddress;
    }
}
exports.HederaToken = HederaToken;
/**
 * The Algo network supports tokens (assets)
 * Algo tokens work similar to native ALGO coin, but the token name is determined by
 * unique asset id on the chain. Internally, BitGo uses token identifiers of the format: (t)algo:<assetId>
 *
 */
class AlgoCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
    }
}
exports.AlgoCoin = AlgoCoin;
/**
 * The Eos network supports tokens
 * Eos tokens work similar to native Eos coin, but the token name is determined by
 * the contractName on the chain.
 *
 */
class EosCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.contractName = options.contractAddress;
        this.contractAddress = options.contractAddress;
        this.symbol = options.symbol ?? options.name.split(':')[1];
    }
}
exports.EosCoin = EosCoin;
/**
 * The Sol network supports tokens
 * Sol tokens work similar to native SOL coin, but the token name is determined by
 * the tokenAddress on the chain.
 *
 */
class SolCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.tokenAddress = options.contractAddress;
        this.contractAddress = options.contractAddress;
        this.programId = options.programId;
    }
}
exports.SolCoin = SolCoin;
class EthLikeERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.EthLikeERC20Token = EthLikeERC20Token;
class EthLikeERC721Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.EthLikeERC721Token = EthLikeERC721Token;
/**
 * The AVAX C Chain network support tokens
 * AVAX C Chain Tokens are ERC20 coins
 */
class AvaxERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.AvaxERC20Token = AvaxERC20Token;
/**
 * The Polygon Chain network support tokens
 * Polygon Chain Tokens are ERC20 coins
 */
class PolygonERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.PolygonERC20Token = PolygonERC20Token;
/**
 * The Arbitrum Chain network support tokens
 * Arbitrum Chain Tokens are ERC20 tokens
 */
class ArbethERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.ArbethERC20Token = ArbethERC20Token;
/**
 * The Optimism Chain network support tokens
 * Optimism Chain Tokens are ERC20 tokens
 */
class OpethERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.OpethERC20Token = OpethERC20Token;
/**
 * The zkSync network support tokens
 * zkSync Tokens are ERC20 tokens
 */
class ZkethERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.ZkethERC20Token = ZkethERC20Token;
/**
 * The Bera Chain network support tokens
 * Bera Chain Tokens are ERC20 tokens
 */
class BeraERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.BeraERC20Token = BeraERC20Token;
/**
 * The Coredao Chain network support tokens
 * Coredao Chain Tokens are ERC20 tokens
 */
class CoredaoERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.CoredaoERC20Token = CoredaoERC20Token;
/**
 * The World Chain network supports tokens
 * World Chain Tokens are ERC20 tokens
 */
class WorldERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.WorldERC20Token = WorldERC20Token;
/**
 * The Flr Chain network supports tokens
 * Flr Chain Tokens are ERC20 tokens
 */
class FlrERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.FlrERC20Token = FlrERC20Token;
/**
 * The XDC network supports tokens
 * XDC Tokens are ERC20 tokens
 */
class XdcERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.XdcERC20Token = XdcERC20Token;
/**
 * The Mon network supports tokens
 * Mon Tokens are ERC20 tokens
 */
class MonERC20Token extends ContractAddressDefinedToken {
    constructor(options) {
        super(options);
    }
}
exports.MonERC20Token = MonERC20Token;
/**
 * The Xrp network supports tokens
 * Xrp tokens are identified by their issuer address
 * Naming format is similar to XLM
 * <network>:<token>-<issuer>
 */
class XrpCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        if (options.domain !== '' && !options.domain.match(constants_1.DOMAIN_PATTERN)) {
            throw new errors_1.InvalidDomainError(options.name, options.domain);
        }
        this.domain = options.domain;
        this.issuerAddress = options.contractAddress.split('::')[0];
        this.currencyCode = options.contractAddress.split('::')[1];
        this.contractAddress = options.contractAddress;
    }
}
exports.XrpCoin = XrpCoin;
class SuiCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.packageId = options.packageId;
        this.module = options.module;
        this.symbol = options.symbol;
        this.contractAddress = options.contractAddress;
    }
}
exports.SuiCoin = SuiCoin;
/**
 * The Apt network supports tokens
 * Apt tokens work similar to native Apt coin, but the token name is determined by
 * the tokenAddress on the chain.
 *
 */
class AptCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.assetId = options.assetId;
    }
}
exports.AptCoin = AptCoin;
/**
 * The Apt network supports non-fungible tokens (Digital Asset Standard)
 * Every NFT belongs to an NFT collection.
 */
class AptNFTCollection extends NFTCollectionIdDefinedToken {
}
exports.AptNFTCollection = AptNFTCollection;
/**
 * The Vechain network supports non-fungible tokens
 * Every NFT belongs to an NFT collection(contract).
 */
class VetNFTCollection extends NFTCollectionIdDefinedToken {
    constructor(options) {
        super(options);
        this.gasTankToken = options.gasTankToken;
    }
}
exports.VetNFTCollection = VetNFTCollection;
/**
 * Fiat currencies, such as USD, EUR, or YEN.
 */
class FiatCoin extends base_1.BaseCoin {
    constructor(options) {
        super({ ...options, kind: base_1.CoinKind.FIAT });
        this.network = options.network;
    }
    requiredFeatures() {
        return new Set([base_1.CoinFeature.ACCOUNT_MODEL]);
    }
    disallowedFeatures() {
        return new Set([base_1.CoinFeature.UNSPENT_MODEL]);
    }
}
exports.FiatCoin = FiatCoin;
FiatCoin.DEFAULT_FEATURES = [...AccountCoin.DEFAULT_FEATURES];
/**
 * The Stacks network supports tokens
 * Stx tokens work similar to native Stx coin, but the token name is determined by
 * the contractName on the chain.
 */
class Sip10Token extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.assetId = options.assetId;
    }
}
exports.Sip10Token = Sip10Token;
/**
 * The Near network supports tokens
 * Near tokens work similar to native near coin
 */
class Nep141Token extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.contractAddress = options.contractAddress;
        this.storageDepositAmount = options.storageDepositAmount;
    }
}
exports.Nep141Token = Nep141Token;
class VetToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.contractAddress = options.contractAddress;
        this.gasTankToken = options.gasTankToken;
    }
}
exports.VetToken = VetToken;
/**
 * Cosmos network supports tokens
 * Cosmos tokens work similar to native coins, but the token is determined by
 * the denom on chain.
 *
 */
class CosmosChainToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.denom = options.denom;
    }
}
exports.CosmosChainToken = CosmosChainToken;
/**
 * TON supports tokens and Jetton is a token standard on TON
 * Jetton tokens work similar to native TON coin
 */
class JettonToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.contractAddress = options.contractAddress;
    }
}
exports.JettonToken = JettonToken;
/**
 * The Bittensor network supports tokens
 * The token name is determined by the subnetId on chain.
 */
class TaoCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.subnetId = options.subnetId;
    }
}
exports.TaoCoin = TaoCoin;
/**
 * The Bittensor network supports tokens
 * The token name is determined by the subnetId on chain.
 */
class PolyxCoin extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.ticker = options.ticker;
        this.assetId = options.assetId;
    }
}
exports.PolyxCoin = PolyxCoin;
class AdaToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.policyId = options.policyId;
        this.assetName = options.assetName;
        this.contractAddress = options.contractAddress;
    }
}
exports.AdaToken = AdaToken;
/**
 * The canton network supports tokens
 * Canton tokens work similar to Canton coin,
 * but they will have their own token standard base url and instrument admin
 */
class CantonToken extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.baseUrl = options.baseUrl;
        this.contractAddress = options.contractAddress;
    }
}
exports.CantonToken = CantonToken;
/**
 * The Tempo network supports TIP20 tokens
 * TIP20 tokens are ERC20-compatible tokens on the Tempo network
 */
class Tip20Token extends AccountCoinToken {
    constructor(options) {
        super({
            ...options,
        });
        this.contractAddress = options.contractAddress;
    }
}
exports.Tip20Token = Tip20Token;
/**
 * Factory function for account coin instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the coin
 * @param fullName Complete human-readable name of the coin
 * @param network Network object for this coin
 * @param decimalPlaces Number of decimal places this coin supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve? The elliptic curve for this chain/token
 * @param prefix? Optional coin prefix. Defaults to empty string
 * @param suffix? Optional coin suffix. Defaults to coin name.
 * @param isToken? Whether or not this account coin is a token of another coin
 */
function account(id, name, fullName, network, decimalPlaces, asset, baseUnit, features = AccountCoin.DEFAULT_FEATURES, primaryKeyCurve = base_1.KeyCurve.Secp256k1, prefix = '', suffix = name.toUpperCase(), isToken = false) {
    return Object.freeze(new AccountCoin({
        id,
        name,
        fullName,
        network,
        prefix,
        suffix,
        baseUnit,
        features,
        decimalPlaces,
        isToken,
        asset,
        primaryKeyCurve,
    }));
}
/**
 * Factory function for gas tank account coin instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the coin
 * @param fullName Complete human-readable name of the coin
 * @param network Network object for this coin
 * @param decimalPlaces Number of decimal places this coin supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param baseUnit
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 * @param gasTankLowBalanceAlertFactor Low gas tank balance alert threshold = (feeEstimate x gasTankLowBalanceAlertFactor)
 * @param gasTankMinBalanceRecommendationFactor Min gas tank balance recommendation = (feeEstimate x gasTankMinBalanceRecommendationFactor)
 * @param prefix Optional coin prefix. Defaults to empty string
 * @param suffix Optional coin suffix. Defaults to coin name.
 * @param isToken Whether or not this account coin is a token of another coin
 */
function gasTankAccount(id, name, fullName, network, decimalPlaces, asset, baseUnit, features = AccountCoin.DEFAULT_FEATURES, primaryKeyCurve = base_1.KeyCurve.Secp256k1, gasTankLowBalanceAlertFactor = 2, gasTankMinBalanceRecommendationFactor = 10, prefix = '', suffix = name.toUpperCase(), isToken = false, gasTankToken) {
    return Object.freeze(new GasTankAccountCoin({
        id,
        name,
        fullName,
        network,
        prefix,
        suffix,
        baseUnit,
        features,
        decimalPlaces,
        isToken,
        asset,
        primaryKeyCurve,
        gasTankLowBalanceAlertFactor,
        gasTankMinBalanceRecommendationFactor,
        gasTankToken,
    }));
}
/**
 * Factory function for ethLikeErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents
 * @param network Optional token network
 * @param coinNames The parent coin names for mainnet and testnet
 * @param features Features of this coin
 * @param prefix Optional token prefix
 * @param suffix Optional token suffix
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function erc20Token(id, name, fullName, decimalPlaces, contractAddress, asset, network, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new EthLikeERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        decimalPlaces,
        asset,
        features,
        prefix,
        suffix,
        primaryKeyCurve,
        isToken: true,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for erc721 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param contractAddress Contract address of this token
 * @param network network
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix
 * @param suffix Optional token suffix
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function erc721Token(id, name, fullName, contractAddress, network, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new EthLikeERC721Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        decimalPlaces: 0, // ERC721 tokens are non-divisible
        asset: base_1.UnderlyingAsset.ERC721,
        features,
        prefix,
        suffix,
        primaryKeyCurve,
        isToken: true,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for erc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Ethereum main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function erc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.BULK_TRANSACTION], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.ethereum, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new Erc20Coin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for testnet erc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the Kovan test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function terc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.BULK_TRANSACTION], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.kovan) {
    return erc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network);
}
/**
 * Factory function for erc721 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param contractAddress Contract address of this token
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Ethereum main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function erc721(id, name, fullName, contractAddress, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.ethereum, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new Erc721Coin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces: 0,
        asset: base_1.UnderlyingAsset.ERC721,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for testnet erc721 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param contractAddress Contract address of this token
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Hoodi test network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function terc721(id, name, fullName, contractAddress, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.hoodi, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return erc721(id, name, fullName, contractAddress, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for nonstandard token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param contractAddress Contract address of this token
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Ethereum main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function nonstandardToken(id, name, fullName, contractAddress, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.ethereum, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new ContractAddressDefinedToken({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces: 0,
        asset: base_1.UnderlyingAsset.NONSTANDARD,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for erc1155 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param contractAddress Contract address of this token
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Ethereum main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function erc1155(id, name, fullName, contractAddress, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.ethereum, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new Erc1155Coin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces: 0,
        asset: base_1.UnderlyingAsset.ERC1155,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for testnet erc1155 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param contractAddress Contract address of this token
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Hoodi test network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function terc1155(id, name, fullName, contractAddress, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.hoodi, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return erc1155(id, name, fullName, contractAddress, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for ERC20-compatible account coin instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param network Network object for this coin
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function erc20CompatibleAccountCoin(id, name, fullName, network, decimalPlaces, contractAddress, asset, baseUnit, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new Erc20CompatibleAccountCoin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: false,
        primaryKeyCurve,
        baseUnit,
    }));
}
/**
 * Factory function for celo token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin
 * @param prefix ? Optional token prefix. Defaults to empty string
 * @param suffix ? Optional token suffix. Defaults to token name.
 * @param network ? Optional token network. Defaults to CELO main network.
 * @param features ? Features of this coin. Defaults to the DEFAULT_FEATURES excluding CUSTODY feature
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function celoToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = coinFeatures_1.CELO_TOKEN_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.celo, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new CeloCoin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for testnet celo token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features ? Features of this coin. Defaults to the DEFAULT_FEATURES excluding CUSTODY feature
 * @param prefix ? Optional token prefix. Defaults to empty string
 * @param suffix ? Optional token suffix. Defaults to token name.
 * @param network ? Optional token network. Defaults to the testnet CELO network.
 */
function tceloToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = coinFeatures_1.CELO_TOKEN_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.celo) {
    return celoToken(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network);
}
/**
 * Factory function for celo token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to BSC main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function bscToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.bsc, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new BscCoin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.BSC,
    }));
}
/**
 * Factory function for testnet bsc token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet BSC network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function tbscToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.bsc) {
    return bscToken(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network);
}
/**
 * Factory function for Stellar token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param domain Domain of the token issuer (used to access token information from the issuer's stellar.toml file)
 * See https://www.stellar.org/developers/guides/concepts/stellar-toml.html
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Stellar mainnet.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function stellarToken(id, name, fullName, decimalPlaces, asset, domain = '', features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.stellar, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new StellarCoin({
        id,
        name,
        fullName,
        decimalPlaces,
        asset,
        domain,
        features,
        prefix,
        suffix,
        network,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.XLM,
    }));
}
/**
 * Factory function for testnet Stellar token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param domain Domain of the token issuer (used to access token information from the issuer's stellar.toml file)
 * See https://www.stellar.org/developers/guides/concepts/stellar-toml.html
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Stellar testnet.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function tstellarToken(id, name, fullName, decimalPlaces, asset, domain = '', features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.stellar) {
    return stellarToken(id, name, fullName, decimalPlaces, asset, domain, features, prefix, suffix, network);
}
/**
 * Factory function for tron token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to TRON main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tronToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.trx, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new TronErc20Coin({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.TRX,
    }));
}
/**
 * Factory function for testnet tron token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet TRON network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function ttronToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.trx, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return tronToken(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for Hedera coin instances
 *
 * @param id uuid v4
 * @param name unique identifier of the coin
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param nodeAccountId node account Id from which the transaction will be sent
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Hedera mainnet.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function hederaCoin(id, name, fullName, network, decimalPlaces, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new HederaCoin({
        id,
        name,
        fullName,
        decimalPlaces,
        asset,
        nodeAccountId: constants_1.HEDERA_NODE_ACCCOUNT_ID,
        features,
        prefix,
        suffix,
        network,
        isToken: false,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.HBAR,
    }));
}
/**
 * Factory function for Hedera token instances
 *
 * @param id uuid v4
 * @param name unique identifier of the coin
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param nodeAccountId node account Id from which the transaction will be sent
 * @param tokenId The unique identifier of this token
 * @param contractAddress Contract address of this token, same as tokenId
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Hedera mainnet.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function hederaToken(id, name, fullName, network, decimalPlaces, asset, contractAddress, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new HederaToken({
        id,
        name,
        fullName,
        decimalPlaces,
        asset,
        nodeAccountId: constants_1.HEDERA_NODE_ACCCOUNT_ID,
        contractAddress,
        features,
        prefix,
        suffix,
        network,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.HBAR,
    }));
}
/**
 * Factory function for ALGO token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token

 * @param alias (optional) alternative identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * See https://developer.algorand.org/docs/reference/transactions/#url
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to ALGO mainnet.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function algoToken(id, name, alias, fullName, decimalPlaces, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.algorand, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new AlgoCoin({
        id,
        name,
        alias,
        fullName,
        decimalPlaces,
        asset,
        features,
        prefix,
        suffix,
        network,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ALGO,
    }));
}
/**
 * Factory function for testnet ALGO token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param alias (optional) alternative identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * See https://developer.algorand.org/docs/reference/transactions/#url
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Algo testnet.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function talgoToken(id, name, alias, fullName, decimalPlaces, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.algorand) {
    return algoToken(id, name, alias, fullName, decimalPlaces, asset, features, prefix, suffix, network);
}
/**
 * Factory function for eos token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractName Contract address of this token
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to EOS main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function eosToken(id, name, fullName, decimalPlaces, contractName, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), symbol, network = networks_1.Networks.main.eos, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new EosCoin({
        id,
        name,
        fullName,
        network,
        contractName,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.EOS,
        symbol,
    }));
}
/**
 * Factory function for testnet eos token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractName Contract address of this token
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param symbol? token symbol as defined in token contract.
 * @param network? Optional token network. Defaults to the testnet EOS network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function teosToken(id, name, fullName, decimalPlaces, contractName, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), symbol, network = networks_1.Networks.test.eos) {
    return eosToken(id, name, fullName, decimalPlaces, contractName, contractAddress, asset, features, prefix, suffix, symbol, network);
}
/**
 * Factory function for sol token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param tokenAddress Token address of this token
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to SOL main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function solToken(id, name, fullName, decimalPlaces, tokenAddress, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.REQUIRES_RESERVE], programId = ProgramID.TokenProgramId, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.sol, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new SolCoin({
        id,
        name,
        fullName,
        network,
        tokenAddress,
        contractAddress,
        prefix,
        suffix,
        features,
        programId,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.SOL,
    }));
}
/**
 * Factory function for testnet solana token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param tokenAddress Token address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet Solana network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 */
function tsolToken(id, name, fullName, decimalPlaces, tokenAddress, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.REQUIRES_RESERVE], programId = ProgramID.TokenProgramId, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.sol) {
    return solToken(id, name, fullName, decimalPlaces, tokenAddress, contractAddress, asset, features, programId, prefix, suffix, network);
}
/**
 * Factory function for prod cardano token instances.
 *
 * @param id uuid v4
 * @param name Name of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param policyId Policy Id
 * @param assetName Asset name -> Policy ID + Asset name is the unique identifier
 * @param contractAddress Fingerprint of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet Cardano network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function adaToken(id, name, fullName, decimalPlaces, policyId, assetName, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.REQUIRES_RESERVE], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.ada, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new AdaToken({
        id,
        name,
        fullName,
        network,
        policyId,
        assetName,
        prefix,
        suffix,
        features,
        decimalPlaces,
        contractAddress,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ADA,
    }));
}
/**
 * Factory function for testnet cardano token instances.
 *
 * @param id uuid v4
 * @param name Name of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param policyId Policy Id
 * @param assetName Asset name -> Policy ID + Asset name is the unique identifier
 * @param contractAddress Fingerprint of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet Cardano network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 */
function tadaToken(id, name, fullName, decimalPlaces, policyId, assetName, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.REQUIRES_RESERVE], network = networks_1.Networks.test.ada, prefix = '', suffix = name.toUpperCase()) {
    return adaToken(id, name, fullName, decimalPlaces, policyId, assetName, contractAddress, asset, features, prefix, suffix, network);
}
/**
 * Factory function for avaxErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to AvalancheC main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function avaxErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.avalancheC, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new AvaxERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for testnet avaxErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the AvalancheC test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tavaxErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.avalancheC, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return avaxErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for polygonErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Polygon main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function polygonErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.polygon, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new PolygonERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for Amoy testnet polygonErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the Polygon test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tpolygonErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.polygon, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return polygonErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for arbethErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Arbitrum main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function arbethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.arbitrum, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new ArbethERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for Arbitrum Sepolia testnet arbethErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the Arbitrum test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tarbethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.arbitrum, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return arbethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for opethErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Optimism main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function opethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559, base_1.CoinFeature.BULK_TRANSACTION], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.optimism, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new OpethERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for Optimism Sepolia testnet opethErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the Optimism test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function topethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.BULK_TRANSACTION], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.optimism, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return opethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for zkethErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to zkSync mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function zkethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.zkSync, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new ZkethERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for zkSync Sepolia testnet zkethErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the zkSync sepolia test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tzkethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.zkSync, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return zkethErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for beraErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to bera mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function beraErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.bera, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new BeraERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for zkSync Sepolia testnet beraErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the bera test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tberaErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.bera, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return beraErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for CoredaoErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to coredao mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function coredaoErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.coredao, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new CoredaoERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for coredao testnet coredaoErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the coredao test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tcoredaoErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.coredao, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return coredaoErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for WorldErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to World Chain mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function worldErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.world, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new WorldERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for world testnet worldErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the World Chain test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tworldErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.world, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return worldErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for FlrErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Flr Chain mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function flrErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.flr, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new FlrERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for Flr testnet FlrErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the Flr Chain test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tflrErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.flr, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return flrErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for XdcErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to XDC mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function xdcErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.xdc, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new XdcERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for XDC testnet XdcErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the XDC test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function txdcErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.xdc, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return xdcErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for MonErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Mon mainnet network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function monErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = [...AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559], prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.mon, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new MonERC20Token({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for Mon testnet MonErc20 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the Mon test network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tmonErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.mon, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return monErc20(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for xrp token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param issuerAddress: The address of the issuer of the token,
 * @param currencyCode The token symbol. Example: USD, BTC, ETH, etc.
 * @param contractAddress Contract address of this token formed with `issuerAddress::currencyCode`
 * @param domain? the domain of the issuer of the token,
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Cardano main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function xrpToken(id, name, fullName, decimalPlaces, issuerAddress, currencyCode, contractAddress, domain = '', asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.xrp, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new XrpCoin({
        id,
        name,
        fullName,
        network,
        issuerAddress,
        currencyCode,
        contractAddress,
        domain,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.XRP,
    }));
}
/**
 * Factory function for testnet cardano token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param issuerAddress: The address of the issuer of the token,
 * @param currencyCode The token symbol. Example: USD, BTC, ETH, etc.
 * @param contractAddress Contract address of this token formed with `issuerAddress::currencyCode`
 * @param domain? the domain of the issuer of the token,
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet Cardano network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function txrpToken(id, name, fullName, decimalPlaces, issuerAddress, currencyCode, contractAddress, domain = '', asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.xrp) {
    return xrpToken(id, name, fullName, decimalPlaces, issuerAddress, currencyCode, contractAddress, domain, asset, features, prefix, suffix, network);
}
/**
 * Factory function for sui token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param packageId PackageId of this token
 * @param module The module of the package with id `packageId`
 * @param symbol Identifies the coin defined in the module `module` of the package with id `packageId`
 * @param contractAddress Contract address of this token formed with `packageId::module::symbol`
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to SUI main network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function suiToken(id, name, fullName, decimalPlaces, packageId, module, symbol, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.sui, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new SuiCoin({
        id,
        name,
        fullName,
        network,
        packageId,
        module,
        symbol,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.SUI,
    }));
}
/**
 * Factory function for testnet sui token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param packageId PackageId of this token
 * @param module The module of the package with id `packageId`
 * @param symbol Identifies the coin defined in the module `module` of the package with id `packageId`
 * @param contractAddress Contract address of this token formed with `packageId::module::symbol`
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to SUI test network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tsuiToken(id, name, fullName, decimalPlaces, packageId, module, symbol, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.sui, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return suiToken(id, name, fullName, decimalPlaces, packageId, module, symbol, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for apt token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param assetId Asset Id of this token i.e. the unique identifier of the token for all tokens - fungible, non-fungible and legacy
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to APT main network.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function aptToken(id, name, fullName, decimalPlaces, assetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.apt, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new AptCoin({
        id,
        name,
        fullName,
        network,
        assetId,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.APT,
    }));
}
/**
 * Factory function for Apt NFT collections.
 *
 * @param id uuid v4
 * @param name unique identifier of the NFT collection
 * @param fullName Complete human-readable name of the NFT collection
 * @param nftCollectionId collection ID of the non-fungible tokens (NFTs)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to APT main network.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function aptNFTCollection(id, name, fullName, nftCollectionId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.apt, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new AptNFTCollection({
        id,
        name,
        fullName,
        network,
        nftCollectionId,
        prefix,
        suffix,
        features,
        decimalPlaces: 0,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.APT,
    }));
}
/**
 * Factory function for testnet apt token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param assetId Asset Id of this token i.e. the unique identifier of the token for all tokens - fungible, non-fungible and legacy
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the testnet APT network.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function taptToken(id, name, fullName, decimalPlaces, assetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.apt, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return aptToken(id, name, fullName, decimalPlaces, assetId, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for testnet Apt NFT collections.
 *
 * @param id uuid v4
 * @param name unique identifier of the NFT collection
 * @param fullName Complete human-readable name of the NFT collection
 * @param nftCollectionId collection ID of the non-fungible tokens (NFTs)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to APT test network.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function taptNFTCollection(id, name, fullName, nftCollectionId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = 't', suffix = name.toUpperCase(), network = networks_1.Networks.test.apt, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return aptNFTCollection(id, name, fullName, nftCollectionId, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for fiat coin instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the coin, should start with 'fiat' or 'tfiat' followed by the 3-char ISO-4217 alphabetical code
 * @param fullName Complete human-readable name of the coin
 * @param network Network object for this coin
 * @param decimalPlaces Number of decimal places this coin supports (divisibility exponent)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `FiatCoin`
 * @param primaryKeyCurve? The elliptic curve for this chain/token
 * @param prefix? Optional coin prefix. Defaults to empty string
 * @param suffix? Optional coin suffix. Defaults to coin name.
 * @param isToken? Whether or not this coin is a token of another coin
 */
function fiat(id, name, fullName, network, decimalPlaces, asset, features = FiatCoin.DEFAULT_FEATURES, primaryKeyCurve = base_1.KeyCurve.Secp256k1, prefix = '', suffix = name.toUpperCase(), isToken = false) {
    return Object.freeze(new FiatCoin({
        id,
        name,
        fullName,
        network,
        prefix,
        suffix,
        features,
        decimalPlaces,
        isToken,
        asset,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.FIAT,
    }));
}
/**
 * Factory function for sip10 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param assetId A unique identifier for a token, which is in the form of contractAddress.contractName::tokenName
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Stacks main network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function sip10Token(id, name, fullName, decimalPlaces, assetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.stx, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new Sip10Token({
        id,
        name,
        fullName,
        network,
        assetId,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.STX,
    }));
}
/**
 * Factory function for testnet sip10 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param assetId A unique identifier for a token, which is in the form of contractAddress.contractName::tokenName
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to the testnet Stacks network.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 */
function tsip10Token(id, name, fullName, decimalPlaces, assetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.stx) {
    return sip10Token(id, name, fullName, decimalPlaces, assetId, asset, features, prefix, suffix, network);
}
/**
 * Factory function for nep141 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param storageDepositAmount the deposit amount needed to get registered with this contract
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to Near main network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function nep141Token(id, name, fullName, decimalPlaces, contractAddress, storageDepositAmount, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.near, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new Nep141Token({
        id,
        name,
        fullName,
        network,
        decimalPlaces,
        contractAddress,
        storageDepositAmount,
        prefix,
        suffix,
        features,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.NEAR,
    }));
}
/**
 * Factory function for testnet nep141 token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param storageDepositAmount the deposit amount needed to get registered with this contract
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the testnet Near network.
 */
function tnep141Token(id, name, fullName, decimalPlaces, contractAddress, storageDepositAmount, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.near) {
    return nep141Token(id, name, fullName, decimalPlaces, contractAddress, storageDepositAmount, asset, features, prefix, suffix, network);
}
/**
 * Factory function for vet token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to Near main network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 * @param gasTankToken representing the token with which gas is paid on the network, defaults to 'VET:VTHO'
 */
function vetToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.vet, primaryKeyCurve = base_1.KeyCurve.Secp256k1, gasTankToken = 'VET:VTHO') {
    return Object.freeze(new VetToken({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.VET,
        gasTankToken,
    }));
}
/**
 * Factory function for testnet vet token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the testnet Near network.
 * @param gasTankToken representing the token with which gas is paid on the network, defaults to 'TVET:VTHO'
 */
function tvetToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.vet, gasTankToken = 'TVET:VTHO' // For test environment
) {
    return vetToken(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, base_1.KeyCurve.Secp256k1, gasTankToken);
}
/**
 * Factory function for Vet NFT collections.
 *
 * @param id uuid v4
 * @param name unique identifier of the NFT collection
 * @param fullName Complete human-readable name of the NFT collection
 * @param nftCollectionId collection ID of the non-fungible tokens (NFTs)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to VET main network.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 * @param gasTankToken representing the token with which gas is paid on the network, defaults to 'VET:VTHO'
 */
function vetNFTCollection(id, name, fullName, nftCollectionId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.vet, primaryKeyCurve = base_1.KeyCurve.Secp256k1, gasTankToken = 'VET:VTHO') {
    return Object.freeze(new VetNFTCollection({
        id,
        name,
        fullName,
        network,
        nftCollectionId,
        prefix,
        suffix,
        features,
        decimalPlaces: 0,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.VET,
        gasTankToken,
    }));
}
/**
 * Factory function for testnet Vet NFT collections.
 *
 * @param id uuid v4
 * @param name unique identifier of the NFT collection
 * @param fullName Complete human-readable name of the NFT collection
 * @param nftCollectionId collection ID of the non-fungible tokens (NFTs)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to VET test network.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES and REQUIRES_RESERVE defined in `AccountCoin`
 * @param primaryKeyCurve The elliptic curve for this chain/token
 * @param gasTankToken representing the token with which gas is paid on the network, defaults to 'TVET:VTHO'
 */
function tvetNFTCollection(id, name, fullName, nftCollectionId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = 't', suffix = name.toUpperCase(), network = networks_1.Networks.test.vet, primaryKeyCurve = base_1.KeyCurve.Secp256k1, gasTankToken = 'TVET:VTHO') {
    return vetNFTCollection(id, name, fullName, nftCollectionId, asset, features, prefix, suffix, network, primaryKeyCurve, gasTankToken);
}
/**
 * Factory function for Cosmos chain token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param denom denomination of this token which will act as a unique identifier for the token on chain
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param network Network (mainnet or testnet) for this token
 * @param baseUnit Base unit of this token (native asset)
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the COSMOS_SIDECHAIN_FEATURES defined in `coinFeatures.ts`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function cosmosToken(id, name, fullName, denom, decimalPlaces, network, baseUnit, asset, features = coinFeatures_1.COSMOS_SIDECHAIN_FEATURES, prefix = '', suffix = name.toUpperCase(), primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new CosmosChainToken({
        id,
        name,
        fullName,
        denom,
        decimalPlaces,
        network,
        baseUnit,
        asset,
        features,
        prefix,
        suffix,
        primaryKeyCurve,
        isToken: true,
    }));
}
/**
 * Factory function for tao token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param subnetId The uid of the subnet this token belongs to, numerical string
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to TAO main network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function taoToken(id, name, fullName, decimalPlaces, subnetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.tao, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new TaoCoin({
        id,
        name,
        fullName,
        network,
        subnetId,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.TAO,
    }));
}
/**
 * Factory function for testnet tao token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param subnetId The uid of the subnet this token belongs to, numerical string
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to TAO test network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function ttaoToken(id, name, fullName, decimalPlaces, subnetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.tao, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return taoToken(id, name, fullName, decimalPlaces, subnetId, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for polyx token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param subnetId The uid of the subnet this token belongs to, numerical string
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Polyx main network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function polyxToken(id, name, fullName, decimalPlaces, ticker, assetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.polyx, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new PolyxCoin({
        id,
        name,
        fullName,
        network,
        ticker,
        assetId,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.POLYX,
    }));
}
/**
 * Factory function for testnet polyx token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param ticker The ticker symbol of the token
 * @param assetId The unique asset identifier for this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES
 * @param prefix? Optional token prefix. Defaults to empty string
 * @param suffix? Optional token suffix. Defaults to token name.
 * @param network? Optional token network. Defaults to Polyx test network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tpolyxToken(id, name, fullName, decimalPlaces, ticker, assetId, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.polyx, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return polyxToken(id, name, fullName, decimalPlaces, ticker, assetId, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for Jetton token instances.
 * Jetton is the token standard on TON blockchain.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to TON main network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function jettonToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.ton, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new JettonToken({
        id,
        name,
        fullName,
        network,
        contractAddress,
        prefix,
        suffix,
        features,
        decimalPlaces,
        asset,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.TON,
    }));
}
/**
 * Factory function for testnet jetton token instances.
 * Jetton is the token standard on TON blockchain.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress Contract address of this token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the DEFAULT_FEATURES defined in `AccountCoin`
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the testnet TON network.
 */
function tjettonToken(id, name, fullName, decimalPlaces, contractAddress, asset, features = AccountCoin.DEFAULT_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.ton, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return jettonToken(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for prod canton token instances.
 *
 * @param id uuid v4
 * @param name Name of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param baseUrl token standard base url, ref: https://docs.digitalasset.com/utilities/devnet/overview/registry-user-guide/token-standard.html#token-standard-endpoints
 * @param contractAddress the contract address of the token in the form `instrumentAdmin:instrumentId`
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to CANTON_FEATURES
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the mainnet canton network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function cantonToken(id, name, fullName, decimalPlaces, baseUrl, contractAddress, asset, features = coinFeatures_1.CANTON_TOKEN_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.canton, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return Object.freeze(new CantonToken({
        id,
        name,
        fullName,
        decimalPlaces,
        network,
        baseUrl,
        contractAddress,
        asset,
        features,
        prefix,
        suffix,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.CANTON,
    }));
}
/**
 * Factory function for testnet canton token instances.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param baseUrl token standard base url
 * @param contractAddress the contract address of the token in the form `instrumentAdmin:instrumentId`
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the CANTON_FEATURES
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the testnet Canton network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tcantonToken(id, name, fullName, decimalPlaces, baseUrl, contractAddress, asset, features = coinFeatures_1.CANTON_TOKEN_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.canton, primaryKeyCurve = base_1.KeyCurve.Ed25519) {
    return cantonToken(id, name, fullName, decimalPlaces, baseUrl, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
/**
 * Factory function for TIP20 token instances on Tempo network.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress the contract address of the token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the TEMPO_FEATURES
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the mainnet Tempo network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function tip20Token(id, name, fullName, decimalPlaces, contractAddress, asset, features = coinFeatures_1.TEMPO_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.main.tempo, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return Object.freeze(new Tip20Token({
        id,
        name,
        fullName,
        decimalPlaces,
        network,
        contractAddress,
        asset,
        features,
        prefix,
        suffix,
        isToken: true,
        primaryKeyCurve,
        baseUnit: base_1.BaseUnit.ETH,
    }));
}
/**
 * Factory function for testnet TIP20 token instances on Tempo network.
 *
 * @param id uuid v4
 * @param name unique identifier of the token
 * @param fullName Complete human-readable name of the token
 * @param decimalPlaces Number of decimal places this token supports (divisibility exponent)
 * @param contractAddress the contract address of the token
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features Features of this coin. Defaults to the TEMPO_FEATURES
 * @param prefix Optional token prefix. Defaults to empty string
 * @param suffix Optional token suffix. Defaults to token name.
 * @param network Optional token network. Defaults to the testnet Tempo network.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function ttip20Token(id, name, fullName, decimalPlaces, contractAddress, asset, features = coinFeatures_1.TEMPO_FEATURES, prefix = '', suffix = name.toUpperCase(), network = networks_1.Networks.test.tempo, primaryKeyCurve = base_1.KeyCurve.Secp256k1) {
    return tip20Token(id, name, fullName, decimalPlaces, contractAddress, asset, features, prefix, suffix, network, primaryKeyCurve);
}
