// @bitgo/statics/dist/src/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCoin = exports.BaseUnit = exports.KeyCurve = exports.UnderlyingAsset = exports.CoinFeature = exports.CoinFamily = exports.CoinKind = void 0;
const errors_1 = require("./errors");
var CoinKind;
(function (CoinKind) {
    CoinKind["CRYPTO"] = "crypto";
    CoinKind["FIAT"] = "fiat";
})(CoinKind || (exports.CoinKind = CoinKind = {}));
/**
 * The coin family links related variants of a single coin together.
 *
 * Typically, each coin will have a testnet and mainnet variant,
 * and these will both have the same coin family.
 *
 * For example, the coins `btc` and `tbtc` both belong to the same family, `btc`.
 */
var CoinFamily;
(function (CoinFamily) {
    CoinFamily["ADA"] = "ada";
    CoinFamily["APECHAIN"] = "apechain";
    CoinFamily["ALGO"] = "algo";
    CoinFamily["APT"] = "apt";
    CoinFamily["ARBETH"] = "arbeth";
    CoinFamily["ASI"] = "asi";
    CoinFamily["ATOM"] = "atom";
    CoinFamily["AVAXC"] = "avaxc";
    CoinFamily["AVAXP"] = "avaxp";
    CoinFamily["BASEETH"] = "baseeth";
    CoinFamily["BABY"] = "baby";
    CoinFamily["BCH"] = "bch";
    CoinFamily["BCHA"] = "bcha";
    CoinFamily["BERA"] = "bera";
    CoinFamily["BLD"] = "bld";
    CoinFamily["BSC"] = "bsc";
    CoinFamily["BSV"] = "bsv";
    CoinFamily["BTC"] = "btc";
    CoinFamily["BTG"] = "btg";
    CoinFamily["CANTON"] = "canton";
    CoinFamily["CELO"] = "celo";
    CoinFamily["CHILIZ"] = "chiliz";
    CoinFamily["CODEXETH"] = "codexeth";
    CoinFamily["COREDAO"] = "coredao";
    CoinFamily["COREUM"] = "coreum";
    CoinFamily["CRONOS"] = "cronos";
    CoinFamily["CSPR"] = "cspr";
    CoinFamily["DASH"] = "dash";
    CoinFamily["DOGE"] = "doge";
    CoinFamily["DOGEOS"] = "dogeos";
    CoinFamily["DOT"] = "dot";
    CoinFamily["DYDX"] = "dydx";
    CoinFamily["EOS"] = "eos";
    CoinFamily["ETC"] = "etc";
    CoinFamily["ETH"] = "eth";
    CoinFamily["ETH2"] = "eth2";
    CoinFamily["ETHW"] = "ethw";
    CoinFamily["FETCHAI"] = "fetchai";
    CoinFamily["FIAT"] = "fiat";
    CoinFamily["FLOW"] = "flow";
    CoinFamily["FANTOM"] = "fantom";
    CoinFamily["FLR"] = "flr";
    CoinFamily["FLRP"] = "flrp";
    CoinFamily["H"] = "h";
    CoinFamily["HASH"] = "hash";
    CoinFamily["HBAR"] = "hbar";
    CoinFamily["HBAREVM"] = "hbarevm";
    CoinFamily["HEMIETH"] = "hemieth";
    CoinFamily["HOODETH"] = "hoodeth";
    CoinFamily["HPPETH"] = "hppeth";
    CoinFamily["ICP"] = "icp";
    CoinFamily["INITIA"] = "initia";
    CoinFamily["INJECTIVE"] = "injective";
    CoinFamily["IOTA"] = "iota";
    CoinFamily["IRYS"] = "irys";
    CoinFamily["ISLM"] = "islm";
    CoinFamily["JOVAYETH"] = "jovayeth";
    CoinFamily["KAIA"] = "kaia";
    CoinFamily["KAVACOSMOS"] = "kavacosmos";
    CoinFamily["KAVAEVM"] = "kavaevm";
    CoinFamily["LNBTC"] = "lnbtc";
    CoinFamily["LTC"] = "ltc";
    CoinFamily["MANTLE"] = "mantle";
    CoinFamily["MANTRA"] = "mantra";
    CoinFamily["MEGAETH"] = "megaeth";
    CoinFamily["MON"] = "mon";
    CoinFamily["XPL"] = "xpl";
    CoinFamily["POLYGON"] = "polygon";
    CoinFamily["POLYX"] = "polyx";
    CoinFamily["PHRS"] = "phrs";
    CoinFamily["CTC"] = "ctc";
    CoinFamily["HYPEEVM"] = "hypeevm";
    CoinFamily["NEAR"] = "near";
    CoinFamily["OAS"] = "oas";
    CoinFamily["OFC"] = "ofc";
    CoinFamily["OG"] = "og";
    CoinFamily["OPBNB"] = "opbnb";
    CoinFamily["OKBXLAYER"] = "okbxlayer";
    CoinFamily["OPETH"] = "opeth";
    CoinFamily["OSMO"] = "osmo";
    CoinFamily["PLUME"] = "plume";
    CoinFamily["RBTC"] = "rbtc";
    CoinFamily["SGB"] = "sgb";
    CoinFamily["SEI"] = "sei";
    CoinFamily["SEIEVM"] = "seievm";
    CoinFamily["SOL"] = "sol";
    CoinFamily["SONIC"] = "sonic";
    CoinFamily["SONEIUM"] = "soneium";
    CoinFamily["STT"] = "stt";
    CoinFamily["SUI"] = "sui";
    CoinFamily["STX"] = "stx";
    CoinFamily["SUSD"] = "susd";
    CoinFamily["TAO"] = "tao";
    CoinFamily["THOR"] = "thor";
    CoinFamily["TIA"] = "tia";
    CoinFamily["TON"] = "ton";
    CoinFamily["TRX"] = "trx";
    CoinFamily["USDT0"] = "usdt0";
    CoinFamily["VET"] = "vet";
    CoinFamily["WORLD"] = "world";
    CoinFamily["WEMIX"] = "wemix";
    CoinFamily["XDC"] = "xdc";
    CoinFamily["XLM"] = "xlm";
    CoinFamily["XRP"] = "xrp";
    CoinFamily["XTZ"] = "xtz";
    CoinFamily["XTZEVM"] = "xtzevm";
    CoinFamily["ZEC"] = "zec";
    CoinFamily["ZETA"] = "zeta";
    CoinFamily["ZKETH"] = "zketh";
    CoinFamily["ZKSYNCERA"] = "zksyncera";
    CoinFamily["LINEAETH"] = "lineaeth";
    CoinFamily["IP"] = "ip";
    CoinFamily["SOMI"] = "somi";
    CoinFamily["FLUENTETH"] = "fluenteth";
    CoinFamily["MORPH"] = "morph";
    CoinFamily["MORPHETH"] = "morpheth";
    CoinFamily["ARCUSDC"] = "arcusdc";
    CoinFamily["TEMPO"] = "tempo";
    CoinFamily["UNIETH"] = "unieth";
})(CoinFamily || (exports.CoinFamily = CoinFamily = {}));
/**
 * Coin features are yes or no questions about what a coin requires or is capable of.
 *
 * This allows coin-agnostic handling of coin-specific features. This is designed
 * to replace checking the coin name against a whitelist of supported coins
 * before executing some coin-specific logic, and instead allows one to check if a
 * coin supports the coin-specific feature that the logic implements.
 */
var CoinFeature;
(function (CoinFeature) {
    /*
     * This coin supports creating wallets on different networks with the same keys. Only works for TSS account-base coins
     */
    CoinFeature["EVM_WALLET"] = "evm-wallet";
    /*
     * This coin supports creating an EVM transaction using Metamask Institutional (MMI).
     */
    CoinFeature["METAMASK_INSTITUTIONAL"] = "metamask-institutional";
    /*
     * The valueless transfer feature indicates that it is valid to send a transaction which moves zero units of the coin.
     *
     * An example is Ethereum, which uses zero value transactions to trigger contract calls.
     */
    CoinFeature["VALUELESS_TRANSFER"] = "valueless-transfer";
    /*
     * Transaction data means there can be arbitrary data encoded in a transaction.
     *
     * Ethereum contract call data is an example.
     */
    CoinFeature["TRANSACTION_DATA"] = "transaction-data";
    /*
     * Some coins have a higher precision range than IEEE 754 doubles, which are used to represent numbers in javascript.
     *
     * For these coins, we must use an arbitrary precision arithmetic library, and this feature indicates this requirement.
     */
    CoinFeature["REQUIRES_BIG_NUMBER"] = "requires-big-number";
    /*
     * RMG requires all wallets to have a backup key held by a BitGo approved Key Recovery Service (KRS)
     */
    CoinFeature["REQUIRES_KRS_BACKUP_KEY"] = "requires-krs-backup-key";
    /*
     * For customers which are not on a postpaid contract, we add an extra output to transactions which pays BitGo a fee.
     *
     * This fee is known as the "pay-as-you-go fee", or just "paygo" for short.
     *
     * Some coins are unable to create transactions with more than one output, so paygo outputs are not possible for these coins.
     */
    CoinFeature["PAYGO"] = "paygo";
    /*
     * Does this coin align with the unspent model?
     *
     * These are typically Bitcoin and forks of it, such as Litecoin and Bitcoin Cash.
     */
    CoinFeature["UNSPENT_MODEL"] = "unspent-model";
    /*
     * Does this coin align with the Lightning Network model?
     *
     * These are typically Lightning Network on unspent model coins, such as BTC and LBTC.
     */
    CoinFeature["LIGHTNING_MODEL"] = "lightning-model";
    /*
     * Does this coin align with the account model?
     *
     * Examples of this coin type are Ethereum, XRP, and Stellar
     */
    CoinFeature["ACCOUNT_MODEL"] = "account-model";
    /*
     * Does this coin support child-pays-for-parent transactions?
     *
     * These are special types of transactions which can accelerate the confirmation time
     * of another transaction which is stuck in the mempool due to low fees.
     *
     * This is only possible for coins which follow the unspent model (UTXO coins).
     */
    CoinFeature["CHILD_PAYS_FOR_PARENT"] = "cpfp";
    /*
     * Does this coin support tokens? These are distinct assets from the underlying coin, but run on the same network.
     *
     * For example, Ethereum's ERC 20 token standard means that it supports tokens, so it shall have this feature.
     */
    CoinFeature["SUPPORTS_TOKENS"] = "supports-tokens";
    /*
     * Are fees for transactions of this coin paid for by the Enterprise (eg, Enterprise gas tank)?
     */
    CoinFeature["ENTERPRISE_PAYS_FEES"] = "enterprise-pays-fees";
    /*
     * This coin requires that accounts keep a minimum balance as reserve
     */
    CoinFeature["REQUIRES_RESERVE"] = "requires-reserve";
    /**
     * @deprecated This property is no longer valid. Please select the following custody option based on the BitGo org:
     * * CUSTODY_BITGO_TRUST
     * * CUSTODY_BITGO_NEW_YORK
     * * CUSTODY_BITGO_GERMANY
     * * CUSTODY_BITGO_SWITZERLAND
     */
    CoinFeature["CUSTODY"] = "custody";
    /*
    This coin uses TSS for key creation and signing
     */
    CoinFeature["TSS"] = "tss";
    /*
     * This coin supports staking
     */
    CoinFeature["STAKING"] = "staking";
    /*
     * This coin supports liquid staking
     */
    CoinFeature["LIQUID_STAKING"] = "liquid-staking";
    /**
     * This coin is deprecated
     */
    CoinFeature["DEPRECATED"] = "deprecated";
    /**
     * This coin is a dummy object meant to be a placeholder for an unsupported token
     */
    CoinFeature["GENERIC_TOKEN"] = "genericToken";
    /*
     * This coin supports custody in BitGo Trust SD entities
     */
    CoinFeature["CUSTODY_BITGO_TRUST"] = "custody-bitgo-trust";
    /*
     * This coin supports custody in BitGo New York entities
     */
    CoinFeature["CUSTODY_BITGO_NEW_YORK"] = "custody-bitgo-new-york";
    /*
     * This coin supports custody in BitGo Germany entities
     */
    CoinFeature["CUSTODY_BITGO_GERMANY"] = "custody-bitgo-germany";
    /*
     * This coin supports custody in BitGo Switzerland entities
     */
    CoinFeature["CUSTODY_BITGO_SWITZERLAND"] = "custody-bitgo-switzerland";
    /*
     * This coin supports custody in BitGo Switzerland entities
     */
    CoinFeature["CUSTODY_BITGO_FRANKFURT"] = "custody-bitgo-frankfurt";
    /*
     * This coin supports custody in BitGo Singapore entities
     */
    CoinFeature["CUSTODY_BITGO_SINGAPORE"] = "custody-bitgo-singapore";
    /*
     * This coin supports custody in BitGo Sister Trust 1 entities
     */
    CoinFeature["CUSTODY_BITGO_SISTER_TRUST_ONE"] = "custody-bitgo-sister-trust-one";
    /**
     * This coin supports custody in BitGo Korea entities
     */
    CoinFeature["CUSTODY_BITGO_KOREA"] = "custody-bitgo-korea";
    /**
     * This coin supports custody in BitGo Europe ApS entities
     */
    CoinFeature["CUSTODY_BITGO_EUROPE_APS"] = "custody-bitgo-europe-aps";
    /**
     * This coin supports custody in BitGo MENA FZE entities
     */
    CoinFeature["CUSTODY_BITGO_MENA_FZE"] = "custody-bitgo-mena-fze";
    /**
     * This coin supports custody in BitGo Custody MENA FZE entities
     */
    CoinFeature["CUSTODY_BITGO_CUSTODY_MENA_FZE"] = "custody-bitgo-custody-mena-fze";
    /**
     * This coin supports custody in BitGo India entities
     */
    CoinFeature["CUSTODY_BITGO_INDIA"] = "custody-bitgo-india";
    /*
     * This coin has transactions that expire after a certain amount of time.
     */
    CoinFeature["EXPIRING_TRANSACTIONS"] = "expiring-transactions";
    /**
     * This coin supports cold wallets that use a multisig signing protocol
     */
    CoinFeature["MULTISIG_COLD"] = "multisig-cold";
    /**
     * This coin supports cold wallets that use a TSS signing protocol
     */
    CoinFeature["TSS_COLD"] = "tss-cold";
    /**
     * This coin uses sha256 hash function for ECDSA TSS signatures
     */
    CoinFeature["SHA256_WITH_ECDSA_TSS"] = "sha256-with-ecdsa-tss";
    /**
     * This coin is cosmos like coin
     */
    CoinFeature["COSMOS_LIKE_COINS"] = "cosmos_like_coins";
    /**
     * This coin supports the ability to rebuild transactions on custody signing
     */
    CoinFeature["REBUILD_ON_CUSTODY_SIGNING"] = "rebuild-on-custody-signing";
    /**
     * This coin supports higher limit for tx request rebuild, which is 10 by default
     */
    CoinFeature["INCREASED_TX_REQUEST_REBUILD_LIMIT"] = "increased-tx-request-rebuild-limit";
    /**
     * This coin supports bulk transaction creation
     */
    CoinFeature["BULK_TRANSACTION"] = "bulk-transaction";
    /**
     * This coin supports bulk ERC20 token transactions (token batching)
     */
    CoinFeature["ERC20_BULK_TRANSACTION"] = "erc20-bulk-transaction";
    /**
     * This coin supports bulk custody transaction creation
     */
    CoinFeature["CUSTODY_BULK_TRANSACTION"] = "custody-bulk-transaction";
    /**
     * This coin supports distributed custody wallets
     */
    CoinFeature["DISTRIBUTED_CUSTODY"] = "distributed-custody";
    /**
     * This coin supports bulk staking transaction creation
     */
    CoinFeature["BULK_STAKING_TRANSACTION"] = "bulk-staking-transaction";
    /**
     * This coin uses non-packed encoding for transaction data
     */
    CoinFeature["USES_NON_PACKED_ENCODING_FOR_TXDATA"] = "uses-non-packed-encoding-for-txdata";
    /**
     * This coins supports MPCv2 for key creation and signing
     */
    CoinFeature["MPCV2"] = "mpcv2";
    /**
     * This coin supports acceleration or nonce filling txn for stuck transactions for tss wallet
     */
    CoinFeature["STUCK_TRANSACTION_MANAGEMENT_TSS"] = "stuck-transaction-management-tss";
    /**
     * This coin supports acceleration or nonce filling txn for stuck transactions for onchain wallet
     */
    CoinFeature["STUCK_TRANSACTION_MANAGEMENT_ONCHAIN"] = "stuck-transaction-management-onchain";
    /**
     * This coin is onboarded on etheruem rollup chain
     */
    CoinFeature["ETH_ROLLUP_CHAIN"] = "eth-rollup-chain";
    /**
     * This coin supports EIP1559 proposal for transaction fee
     */
    CoinFeature["EIP1559"] = "EIP1559";
    /**
     * Fees for transactions of TSS wallet of this coin would be paid by the Enterprise i.e. Gas Tank
     */
    CoinFeature["TSS_ENTERPRISE_PAYS_FEES"] = "tss-enterprise-pays-fees";
    /**
     * Indicates that fees for transactions on a wallet for this coin are paid with a token (not the native coin).
     */
    CoinFeature["FEES_PAID_WITH_TOKEN"] = "fees-paid-with-token";
    /**
     * This coin supports alphanumeric memo id
     */
    CoinFeature["ALPHANUMERIC_MEMO_ID"] = "alphanumeric-memo-id";
    /**
     * This coin supports WalletConnect
     */
    CoinFeature["WALLET_CONNECT_DEFI"] = "wallet-connect-defi";
    /**
     * This coin is gated for TSS Support
     */
    CoinFeature["TSS_SUPPORT_GATED"] = "tss-support-gated";
    /**
     * This coin is gated for Multisig Support
     */
    CoinFeature["MULTISIG_SUPPORT_GATED"] = "multisig-support-gated";
    /**
     * This coins is an EVM compatible coin and should use common EVM functionality
     */
    CoinFeature["SHARED_EVM_SIGNING"] = "shared-evm-signing";
    /**
     * This coin is an EVM compatible coin and should use common EVM SDK module
     */
    CoinFeature["SHARED_EVM_SDK"] = "shared-evm-sdk";
    /**
     * This coin supports erc20 tokens
     */
    CoinFeature["SUPPORTS_ERC20"] = "supports-erc20-token";
    /**
     * This coin supports erc721 tokens
     */
    CoinFeature["SUPPORTS_ERC721"] = "supports-erc721-token";
    /**
     * This coin is a Cosmos coin and should use shared Cosmos SDK module
     */
    CoinFeature["SHARED_COSMOS_SDK"] = "shared-cosmos-sdk";
    /**
     * This coin is a Cosmos coin and should use shared Cosmos Functionality in WP
     */
    CoinFeature["SHARED_COSMOS_WP"] = "shared-cosmos-wp";
    /**
     * This coin is a Cosmos coin and should use common Cosmos logic in BGA
     */
    CoinFeature["COSMOS_COMMON_BGA"] = "cosmos-common-bga";
    /**
     * This coin is EVM based coin
     */
    CoinFeature["EVM_COIN"] = "evm_coin";
    /**
     * This coin supports multisig wallets
     */
    CoinFeature["MULTISIG"] = "multisig";
    /**
     * This coin is an EVM compatible coin and should use common EVM model registration in IMS
     */
    CoinFeature["EVM_COMPATIBLE_IMS"] = "evm-compatible-ims";
    /**
     * This coin is an EVM compatible coin and should use common EVM logic in UI
     */
    CoinFeature["EVM_COMPATIBLE_UI"] = "evm-compatible-ui";
    /**
     * This coin is an EVM compatible coin which supports unsigned sweep recovery
     */
    CoinFeature["EVM_UNSIGNED_SWEEP_RECOVERY"] = "evm-unsigned-sweep-recovery";
    /**
     * This coin is an EVM compatible coin which supports non-bitgo recovery
     */
    CoinFeature["EVM_NON_BITGO_RECOVERY"] = "evm-non-bitgo-recovery";
    /**
     * This coin is a rebase token and should use the rebase token functionality
     */
    CoinFeature["REBASE_TOKEN"] = "rebase-token";
    /**
     * This coin is an EVM compatible coin and should use common EVM logic in WP
     */
    CoinFeature["EVM_COMPATIBLE_WP"] = "evm-compatible-wp";
    /**
     * This token is internal and shouldn't be exposed to users
     */
    CoinFeature["RESTRICTED"] = "restricted";
    /**
     * This coin is an EVM compatible coin and should use common EVM message signing functionality
     */
    CoinFeature["SHARED_EVM_MESSAGE_SIGNING"] = "shared-evm-message-signing";
    /**
     * This token is a stablecoin
     */
    CoinFeature["STABLECOIN"] = "stablecoin";
    /**
     * This coin supports alternative address identifier format
     */
    CoinFeature["ALTERNATIVE_ADDRESS_IDENTIFIER"] = "alternative-address-identifier";
    /**
     * This token standard uses alternative address identifiers (e.g., DIDs for Polymesh tokens)
     */
    CoinFeature["TOKEN_STANDARD_USES_ALTERNATIVE_ADDRESS_IDENTIFIER"] = "token-standard-uses-alternative-address-identifier";
    /**
     * This coin supports one-step deposit
     */
    CoinFeature["SUPPORTS_ONE_STEP_DEPOSIT"] = "supports-one-step-deposit";
    /**
     * This coin requires a wallet initialization transaction
     */
    CoinFeature["REQUIRES_WALLET_INITIALIZATION_TRANSACTION"] = "requires-wallet-initialization-transaction";
    /**
     * This coin requires a deposit acceptance transaction
     */
    CoinFeature["REQUIRES_DEPOSIT_ACCEPTANCE_TRANSACTION"] = "requires-deposit-acceptance-transaction";
    /**
     * This coin allows negative fees in transactions
     */
    CoinFeature["ALLOWS_NEGATIVE_FEE"] = "allows-negative-fee";
})(CoinFeature || (exports.CoinFeature = CoinFeature = {}));
/**
 * Some coins are representations of another underlying asset class. An example
 * is Wrapped Bitcoin, which represents Bitcoin on the Ethereum blockchain.
 *
 * For these coins, the `UnderlyingAsset` provides a link to the actual
 * asset for which the coin is a unit of account.
 */
var UnderlyingAsset;
(function (UnderlyingAsset) {
    UnderlyingAsset["INVALID_UNKNOWN"] = "invalid_asset_type";
    UnderlyingAsset["ADA"] = "ada";
    UnderlyingAsset["ALGO"] = "algo";
    UnderlyingAsset["APE"] = "ape";
    UnderlyingAsset["APECHAIN"] = "apechain";
    UnderlyingAsset["API3"] = "api3";
    UnderlyingAsset["ARBETH"] = "arbeth";
    UnderlyingAsset["BASEETH"] = "baseeth";
    UnderlyingAsset["ASI"] = "asi";
    UnderlyingAsset["ATOM"] = "atom";
    UnderlyingAsset["AVAXC"] = "avaxc";
    UnderlyingAsset["AVAXP"] = "avaxp";
    UnderlyingAsset["AXL"] = "AXL";
    UnderlyingAsset["AXLV2"] = "axlv2";
    UnderlyingAsset["BABY"] = "baby";
    UnderlyingAsset["BCH"] = "bch";
    UnderlyingAsset["BCHA"] = "bcha";
    UnderlyingAsset["BERA"] = "bera";
    UnderlyingAsset["BLD"] = "bld";
    UnderlyingAsset["BSC"] = "bsc";
    UnderlyingAsset["BSV"] = "bsv";
    UnderlyingAsset["BTC"] = "btc";
    UnderlyingAsset["BTG"] = "btg";
    UnderlyingAsset["CANTON"] = "canton";
    UnderlyingAsset["DASH"] = "dash";
    UnderlyingAsset["DOT"] = "dot";
    UnderlyingAsset["CELO"] = "celo";
    UnderlyingAsset["CHILIZ"] = "chiliz";
    UnderlyingAsset["CODEXETH"] = "codexeth";
    UnderlyingAsset["COREDAO"] = "coredao";
    UnderlyingAsset["COREUM"] = "coreum";
    UnderlyingAsset["CRONOS"] = "cronos";
    UnderlyingAsset["CSPR"] = "cspr";
    UnderlyingAsset["EOS"] = "eos";
    UnderlyingAsset["ERD"] = "erd";
    UnderlyingAsset["ETC"] = "etc";
    UnderlyingAsset["ETH"] = "eth";
    UnderlyingAsset["ETH2"] = "eth2";
    UnderlyingAsset["ETHW"] = "ethw";
    UnderlyingAsset["EURCVV0"] = "eurcvv0";
    UnderlyingAsset["EURCV"] = "eurcv";
    UnderlyingAsset["EUROC"] = "euroc";
    UnderlyingAsset["EURR"] = "eurr";
    UnderlyingAsset["FETCHAI"] = "fetchai";
    UnderlyingAsset["FLOW"] = "flow";
    UnderlyingAsset["FLR"] = "flr";
    UnderlyingAsset["FLRP"] = "flrp";
    UnderlyingAsset["FLUENTETH"] = "fluenteth";
    UnderlyingAsset["FANTOM"] = "fantom";
    UnderlyingAsset["GTC"] = "gtc";
    UnderlyingAsset["H"] = "h";
    UnderlyingAsset["HASH"] = "hash";
    UnderlyingAsset["HBAR"] = "hbar";
    UnderlyingAsset["HBAREVM"] = "hbarevm";
    UnderlyingAsset["HEMIETH"] = "hemieth";
    UnderlyingAsset["HOODETH"] = "hoodeth";
    UnderlyingAsset["HPPETH"] = "hppeth";
    UnderlyingAsset["ICP"] = "icp";
    UnderlyingAsset["IP"] = "ip";
    UnderlyingAsset["INITIA"] = "initia";
    UnderlyingAsset["INJECTIVE"] = "injective";
    UnderlyingAsset["IOTA"] = "iota";
    UnderlyingAsset["IRYS"] = "irys";
    UnderlyingAsset["ISLM"] = "islm";
    UnderlyingAsset["JOVAYETH"] = "jovayeth";
    UnderlyingAsset["KAIA"] = "kaia";
    UnderlyingAsset["KAVACOSMOS"] = "kavacosmos";
    UnderlyingAsset["KAVAEVM"] = "kavaevm";
    UnderlyingAsset["LNBTC"] = "lnbtc";
    UnderlyingAsset["LTC"] = "ltc";
    UnderlyingAsset["LINEAETH"] = "lineaeth";
    UnderlyingAsset["MANTLE"] = "mantle";
    UnderlyingAsset["MANTRA"] = "mantra";
    UnderlyingAsset["MEGAETH"] = "megaeth";
    UnderlyingAsset["MON"] = "mon";
    UnderlyingAsset["MORPH"] = "morph";
    UnderlyingAsset["MORPHETH"] = "morpheth";
    UnderlyingAsset["NEAR"] = "near";
    UnderlyingAsset["OAS"] = "oas";
    UnderlyingAsset["OG"] = "og";
    UnderlyingAsset["OKBXLAYER"] = "okbxlayer";
    UnderlyingAsset["OPBNB"] = "opbnb";
    UnderlyingAsset["OPETH"] = "opeth";
    UnderlyingAsset["OSMO"] = "osmo";
    UnderlyingAsset["XPL"] = "xpl";
    UnderlyingAsset["POLYGON"] = "polygon";
    UnderlyingAsset["PHRS"] = "phrs";
    UnderlyingAsset["PLUME"] = "plume";
    UnderlyingAsset["CTC"] = "ctc";
    UnderlyingAsset["HYPEEVM"] = "hypeevm";
    UnderlyingAsset["RBTC"] = "rbtc";
    UnderlyingAsset["SEI"] = "sei";
    UnderlyingAsset["SEIEVM"] = "seievm";
    UnderlyingAsset["SGB"] = "sgb";
    UnderlyingAsset["SOL"] = "sol";
    UnderlyingAsset["SOMI"] = "somi";
    UnderlyingAsset["SONEIUM"] = "soneium";
    UnderlyingAsset["SONIC"] = "sonic";
    UnderlyingAsset["STT"] = "stt";
    UnderlyingAsset["STX"] = "stx";
    UnderlyingAsset["SUI"] = "sui";
    UnderlyingAsset["TIA"] = "tia";
    UnderlyingAsset["TON"] = "ton";
    UnderlyingAsset["TRX"] = "trx";
    UnderlyingAsset["USDT0"] = "usdt0";
    UnderlyingAsset["VET"] = "vet";
    UnderlyingAsset["WEMIX"] = "wemix";
    UnderlyingAsset["WORLD"] = "world";
    UnderlyingAsset["XDC"] = "xdc";
    UnderlyingAsset["XLM"] = "xlm";
    UnderlyingAsset["XRP"] = "xrp";
    UnderlyingAsset["XTZ"] = "xtz";
    UnderlyingAsset["XTZEVM"] = "xtzevm";
    UnderlyingAsset["ZEC"] = "zec";
    UnderlyingAsset["ZETA"] = "zeta";
    UnderlyingAsset["ZKETH"] = "zketh";
    UnderlyingAsset["ZKSYNCERA"] = "zksyncera";
    UnderlyingAsset["UNIETH"] = "unieth";
    // ERC 20 tokens
    UnderlyingAsset["$Evmosia.com"] = "$evmosia.com";
    UnderlyingAsset["0xREVIEW"] = "0xreview";
    UnderlyingAsset["1INCH"] = "1inch";
    UnderlyingAsset["1UP"] = "1up";
    UnderlyingAsset["3CRV"] = "3crv";
    UnderlyingAsset["AAVE"] = "aave";
    UnderlyingAsset["ABT"] = "abt";
    UnderlyingAsset["ACE"] = "ace";
    UnderlyingAsset["ACEV2"] = "acev2";
    UnderlyingAsset["ACX"] = "acx";
    UnderlyingAsset["ACXT"] = "acxt";
    UnderlyingAsset["ACH"] = "ach";
    UnderlyingAsset["ADABEAR"] = "adabear";
    UnderlyingAsset["ADABULL"] = "adabull";
    UnderlyingAsset["ADX"] = "adx";
    UnderlyingAsset["AE"] = "ae";
    UnderlyingAsset["AERGO"] = "aergo";
    UnderlyingAsset["AERGO1"] = "aergo1";
    UnderlyingAsset["AGEUR"] = "ageur";
    UnderlyingAsset["AGI"] = "agi";
    UnderlyingAsset["AGIX"] = "agix";
    UnderlyingAsset["AGLD"] = "agld";
    UnderlyingAsset["AGWD"] = "agwd";
    UnderlyingAsset["AION"] = "aion";
    UnderlyingAsset["AJNA"] = "ajna";
    UnderlyingAsset["AKRO"] = "akro";
    UnderlyingAsset["ALCX"] = "alcx";
    UnderlyingAsset["ALD"] = "ald";
    UnderlyingAsset["ALDRIN"] = "aldrin";
    UnderlyingAsset["ALEPH"] = "aleph";
    UnderlyingAsset["ALGOBEAR"] = "algobear";
    UnderlyingAsset["ALGOBULL"] = "algobull";
    UnderlyingAsset["ALGODOOM"] = "algodoom";
    UnderlyingAsset["ALGOHEDGE"] = "algohedge";
    UnderlyingAsset["ALGOMOON"] = "algomoon";
    UnderlyingAsset["ALTDOOM"] = "altdoom";
    UnderlyingAsset["ALTMOON"] = "altmoon";
    UnderlyingAsset["ALI"] = "ali";
    UnderlyingAsset["ALICE"] = "alice";
    UnderlyingAsset["ALK"] = "alk";
    UnderlyingAsset["ALM"] = "alm";
    UnderlyingAsset["ALPHA"] = "alpha";
    UnderlyingAsset["ALTBEAR"] = "altbear";
    UnderlyingAsset["ALTBULL"] = "altbull";
    UnderlyingAsset["ALTHEDGE"] = "althedge";
    UnderlyingAsset["AMKT"] = "amkt";
    UnderlyingAsset["AMN"] = "amn";
    UnderlyingAsset["AMO"] = "amo";
    UnderlyingAsset["AMP"] = "amp";
    UnderlyingAsset["AMPL"] = "ampl";
    UnderlyingAsset["AMON"] = "amon";
    UnderlyingAsset["AMPX"] = "ampx";
    UnderlyingAsset["ANA"] = "ana";
    UnderlyingAsset["ANC"] = "anc";
    UnderlyingAsset["ANGLE"] = "angle";
    UnderlyingAsset["ANKR"] = "ankr";
    UnderlyingAsset["ANKRETH"] = "ankreth";
    UnderlyingAsset["ANML"] = "anml";
    UnderlyingAsset["ANT"] = "ant";
    UnderlyingAsset["ANTV2"] = "antv2";
    UnderlyingAsset["AOA"] = "aoa";
    UnderlyingAsset["APPC"] = "appc";
    UnderlyingAsset["APT"] = "apt";
    UnderlyingAsset["AQT"] = "aqt";
    UnderlyingAsset["ARCUSDC"] = "arcusdc";
    UnderlyingAsset["ARCT"] = "arct";
    UnderlyingAsset["ARCX"] = "arcx";
    UnderlyingAsset["ARKM"] = "arkm";
    UnderlyingAsset["ARMOR"] = "armor";
    UnderlyingAsset["ARPA"] = "arpa";
    UnderlyingAsset["ARTEQ"] = "arteq";
    UnderlyingAsset["ASD"] = "asd";
    UnderlyingAsset["AST"] = "ast";
    UnderlyingAsset["ASTO"] = "asto";
    UnderlyingAsset["ATA"] = "ata";
    UnderlyingAsset["ATF"] = "atf";
    UnderlyingAsset["ATH"] = "ath";
    UnderlyingAsset["ATL"] = "atl";
    UnderlyingAsset["ATLAS"] = "atlas";
    UnderlyingAsset["ATOMBEAR"] = "atombear";
    UnderlyingAsset["ATOMBULL"] = "atombull";
    UnderlyingAsset["ATRI"] = "atri";
    UnderlyingAsset["AUCTION"] = "auction";
    UnderlyingAsset["AUDD"] = "audd";
    UnderlyingAsset["AUDF"] = "audf";
    UnderlyingAsset["AUDIO"] = "audio";
    UnderlyingAsset["AUDX"] = "audx";
    UnderlyingAsset["AUSD"] = "ausd";
    UnderlyingAsset["AUSDT"] = "ausdt";
    UnderlyingAsset["AUST"] = "aust";
    UnderlyingAsset["AVA"] = "ava";
    UnderlyingAsset["AVT"] = "avt";
    UnderlyingAsset["AWBTC"] = "awbtc";
    UnderlyingAsset["AXPR"] = "axpr";
    UnderlyingAsset["AXS"] = "axs";
    UnderlyingAsset["AXSV2"] = "axsv2";
    UnderlyingAsset["AYFI"] = "ayfi";
    UnderlyingAsset["AZUKI"] = "azuki";
    UnderlyingAsset["AZUKI2"] = "azuki2";
    UnderlyingAsset["AZUKIPEPE"] = "azukipepe";
    UnderlyingAsset["BADGER"] = "badger";
    UnderlyingAsset["BAI"] = "bai";
    UnderlyingAsset["BAL"] = "bal";
    UnderlyingAsset["BAND"] = "band";
    UnderlyingAsset["BANK"] = "bank";
    UnderlyingAsset["BAO"] = "bao";
    UnderlyingAsset["BASIC"] = "basic";
    UnderlyingAsset["BAT"] = "bat";
    UnderlyingAsset["BAX"] = "bax";
    UnderlyingAsset["BBANK"] = "bbank";
    UnderlyingAsset["BBSAMO"] = "bbsamo";
    UnderlyingAsset["BBTC"] = "BBTC";
    UnderlyingAsset["BBX"] = "bbx";
    UnderlyingAsset["BCAP"] = "bcap";
    UnderlyingAsset["BCC"] = "bcc";
    UnderlyingAsset["BCHBEAR"] = "bchbear";
    UnderlyingAsset["BCHBULL"] = "bchbull";
    UnderlyingAsset["BCHDOOM"] = "bchdoom";
    UnderlyingAsset["BCHHEDGE"] = "bchhedge";
    UnderlyingAsset["BCHMOON"] = "bchmoon";
    UnderlyingAsset["BCIO"] = "bcio";
    UnderlyingAsset["BCUT"] = "bcut";
    UnderlyingAsset["BCT"] = "bct";
    UnderlyingAsset["BDXN"] = "bdxn";
    UnderlyingAsset["BEAM"] = "beam";
    UnderlyingAsset["BEAR"] = "bear";
    UnderlyingAsset["BEARSHIT"] = "bearshit";
    UnderlyingAsset["BED"] = "bed";
    UnderlyingAsset["BEND"] = "bend";
    UnderlyingAsset["BEPRO"] = "bepro";
    UnderlyingAsset["BETA"] = "beta";
    UnderlyingAsset["BGB"] = "bgb";
    UnderlyingAsset["BGBG"] = "bgbg";
    UnderlyingAsset["BICO"] = "bico";
    UnderlyingAsset["BID"] = "bid";
    UnderlyingAsset["BIDL"] = "bidl";
    UnderlyingAsset["BIGTIME"] = "bigtime";
    UnderlyingAsset["BIRD"] = "bird";
    UnderlyingAsset["BIT"] = "bit";
    UnderlyingAsset["BKT"] = "bkt";
    UnderlyingAsset["BKX"] = "bkx";
    UnderlyingAsset["BLCT"] = "blct";
    UnderlyingAsset["BLT"] = "blt";
    UnderlyingAsset["BLUR"] = "blur";
    UnderlyingAsset["BLUR0x083"] = "blur0x083";
    UnderlyingAsset["BLUR0xb93"] = "blur0xb93";
    UnderlyingAsset["BLZ"] = "blz";
    UnderlyingAsset["BNB"] = "bnb";
    UnderlyingAsset["BNBBEAR"] = "bnbbear";
    UnderlyingAsset["BNBBULL"] = "bnbbull";
    UnderlyingAsset["BNBDOOM"] = "bnbdoom";
    UnderlyingAsset["BNBHEDGE"] = "bnbhedge";
    UnderlyingAsset["BNBMOON"] = "bnbmoon";
    UnderlyingAsset["BNK"] = "bnk";
    UnderlyingAsset["BNL"] = "bnl";
    UnderlyingAsset["BNT"] = "bnt";
    UnderlyingAsset["BNTY"] = "bnty";
    UnderlyingAsset["BNVDA"] = "bnvda";
    UnderlyingAsset["BOB"] = "bob";
    UnderlyingAsset["BOND"] = "bond";
    UnderlyingAsset["BONK"] = "bonk";
    UnderlyingAsset["BONE"] = "bone";
    UnderlyingAsset["BORG"] = "borg";
    UnderlyingAsset["BOTTO"] = "botto";
    UnderlyingAsset["BLOCKS"] = "blocks";
    UnderlyingAsset["BOX"] = "box";
    UnderlyingAsset["BOBA"] = "boba";
    UnderlyingAsset["BRD"] = "brd";
    UnderlyingAsset["BRIBE"] = "bribe";
    UnderlyingAsset["BRZ"] = "brz";
    UnderlyingAsset["BSGG"] = "bsgg";
    UnderlyingAsset["BST"] = "bst";
    UnderlyingAsset["BSVBEAR"] = "bsvbear";
    UnderlyingAsset["BSVBULL"] = "bsvbull";
    UnderlyingAsset["BSVDOOM"] = "bsvdoom";
    UnderlyingAsset["BSVHEDGE"] = "bsvhedge";
    UnderlyingAsset["BSVMOON"] = "bsvmoon";
    UnderlyingAsset["BSX"] = "bsx";
    UnderlyingAsset["BTC2XFLI"] = "btc2xfli";
    UnderlyingAsset["BTMXBEAR"] = "btmxbear";
    UnderlyingAsset["BTMXBULL"] = "btmxbull";
    UnderlyingAsset["BTRST"] = "btrst";
    UnderlyingAsset["BTSG"] = "btsg";
    UnderlyingAsset["BTT"] = "btt";
    UnderlyingAsset["BTU"] = "btu";
    UnderlyingAsset["BUIDL"] = "buidl";
    UnderlyingAsset["BULL"] = "bull";
    UnderlyingAsset["BULLSHIT"] = "bullshit";
    UnderlyingAsset["BURP"] = "burp";
    UnderlyingAsset["BUSD"] = "busd";
    UnderlyingAsset["BUY"] = "buy";
    UnderlyingAsset["BPT"] = "bpt";
    UnderlyingAsset["BVOL"] = "bvol";
    UnderlyingAsset["BXX"] = "bxx";
    UnderlyingAsset["BXXV1"] = "bxxv1";
    UnderlyingAsset["BZZ"] = "bzz";
    UnderlyingAsset["C3"] = "c3";
    UnderlyingAsset["C6P"] = "c6p";
    UnderlyingAsset["C8P"] = "c8p";
    UnderlyingAsset["C98"] = "c98";
    UnderlyingAsset["CACXT"] = "cacxt";
    UnderlyingAsset["CADX"] = "cadx";
    UnderlyingAsset["CAG"] = "cag";
    UnderlyingAsset["CANTO"] = "canto";
    UnderlyingAsset["CAPS"] = "caps";
    UnderlyingAsset["CARV"] = "carv";
    UnderlyingAsset["CASH"] = "cash";
    UnderlyingAsset["CBAT"] = "cbat";
    UnderlyingAsset["CBC"] = "cbc";
    UnderlyingAsset["CBETH"] = "cbeth";
    UnderlyingAsset["CBRL"] = "cbrl";
    UnderlyingAsset["CCAI"] = "ccai";
    UnderlyingAsset["CCT"] = "cct";
    UnderlyingAsset["CDAG"] = "cdag";
    UnderlyingAsset["CDAI"] = "cdai";
    UnderlyingAsset["CDAIV2"] = "cdaiV2";
    UnderlyingAsset["CDT"] = "cdt";
    UnderlyingAsset["CEL"] = "cel";
    UnderlyingAsset["CELLS"] = "cells";
    UnderlyingAsset["CELR"] = "celr";
    UnderlyingAsset["CERE"] = "cere";
    UnderlyingAsset["CETH"] = "ceth";
    UnderlyingAsset["CFX"] = "cfx";
    UnderlyingAsset["CHAINLINK"] = "chainlink";
    UnderlyingAsset["CHART"] = "chart";
    UnderlyingAsset["CHO"] = "cho";
    UnderlyingAsset["CHFX"] = "chfx";
    UnderlyingAsset["CHR"] = "chr";
    UnderlyingAsset["CHSB"] = "chsb";
    UnderlyingAsset["CHZ"] = "chz";
    UnderlyingAsset["CIBO"] = "cibo";
    UnderlyingAsset["CIX100"] = "cix100";
    UnderlyingAsset["CLIQ"] = "cliq";
    UnderlyingAsset["CLN"] = "cln";
    UnderlyingAsset["CLT"] = "clt";
    UnderlyingAsset["CLXY"] = "clxy";
    UnderlyingAsset["CLV"] = "clv";
    UnderlyingAsset["CMFI"] = "cmfi";
    UnderlyingAsset["CNFI"] = "cnfi";
    UnderlyingAsset["CNG"] = "cng";
    UnderlyingAsset["CNYX"] = "cnyx";
    UnderlyingAsset["COLLAR"] = "collar";
    UnderlyingAsset["COMBO"] = "combo";
    UnderlyingAsset["COMP"] = "comp";
    UnderlyingAsset["CONV"] = "conv";
    UnderlyingAsset["COPE"] = "cope";
    UnderlyingAsset["CORE"] = "core";
    UnderlyingAsset["COS"] = "cos";
    UnderlyingAsset["COTI"] = "coti";
    UnderlyingAsset["COVAL"] = "coval";
    UnderlyingAsset["COVER"] = "cover";
    UnderlyingAsset["COVERPROTOCOL"] = "coverprotocol";
    UnderlyingAsset["COW"] = "cow";
    UnderlyingAsset["CPAY"] = "cpay";
    UnderlyingAsset["CPLT"] = "cplt";
    UnderlyingAsset["CPOOL"] = "cpool";
    UnderlyingAsset["CQT"] = "cqt";
    UnderlyingAsset["CQX"] = "cqx";
    UnderlyingAsset["CRA"] = "cra";
    UnderlyingAsset["CRDT"] = "crdt";
    UnderlyingAsset["CRE"] = "cre";
    UnderlyingAsset["CREAM"] = "cream";
    UnderlyingAsset["CREP"] = "crep";
    UnderlyingAsset["CRI"] = "cri";
    UnderlyingAsset["CRO"] = "cro";
    UnderlyingAsset["CRV"] = "crv";
    UnderlyingAsset["CRPT"] = "crpt";
    UnderlyingAsset["CRPT1"] = "crpt1";
    UnderlyingAsset["CSLV"] = "cslv";
    UnderlyingAsset["CSOL"] = "csol";
    UnderlyingAsset["CSP"] = "csp";
    UnderlyingAsset["CTSI"] = "ctsi";
    UnderlyingAsset["CTX"] = "ctx";
    UnderlyingAsset["CUBE"] = "cube";
    UnderlyingAsset["CUSD"] = "cusd";
    UnderlyingAsset["CUSDC"] = "cusdc";
    UnderlyingAsset["CVXFXS"] = "cvxfxs";
    UnderlyingAsset["CWAR"] = "cwar";
    UnderlyingAsset["CWBTC"] = "cwbtc";
    UnderlyingAsset["CVC"] = "cvc";
    UnderlyingAsset["CVX"] = "cvx";
    UnderlyingAsset["CXT"] = "cxt";
    UnderlyingAsset["CYBER"] = "cyber";
    UnderlyingAsset["CZRX"] = "czrx";
    UnderlyingAsset["DACXI"] = "dacxi";
    UnderlyingAsset["DADI"] = "dadi";
    UnderlyingAsset["DAMM"] = "damm";
    UnderlyingAsset["DAI"] = "dai";
    UnderlyingAsset["DAO"] = "dao";
    UnderlyingAsset["DAOLANG"] = "daolang";
    UnderlyingAsset["DAR"] = "dar";
    UnderlyingAsset["DATA"] = "data";
    UnderlyingAsset["DATAV2"] = "datav2";
    UnderlyingAsset["DATAECON"] = "dataecon";
    UnderlyingAsset["DAWN"] = "dawn";
    UnderlyingAsset["DEC"] = "dec";
    UnderlyingAsset["DEGO"] = "dego";
    UnderlyingAsset["DENT"] = "dent";
    UnderlyingAsset["DEP"] = "dep";
    UnderlyingAsset["DEPAY"] = "depay";
    UnderlyingAsset["DEXA"] = "dexa";
    UnderlyingAsset["DEXE"] = "dexe";
    UnderlyingAsset["DFD"] = "dfd";
    UnderlyingAsset["DFI"] = "dfi";
    UnderlyingAsset["DFL"] = "dfl";
    UnderlyingAsset["DFX"] = "dfx";
    UnderlyingAsset["DGCL"] = "dgcl";
    UnderlyingAsset["DGD"] = "dgd";
    UnderlyingAsset["DGLD"] = "dgld";
    UnderlyingAsset["DGX"] = "dgx";
    UnderlyingAsset["DHT"] = "dht";
    UnderlyingAsset["DIGG"] = "digg";
    UnderlyingAsset["DIA"] = "dia";
    UnderlyingAsset["DING"] = "ding";
    UnderlyingAsset["DIPE"] = "dipe";
    UnderlyingAsset["DMG"] = "dmg";
    UnderlyingAsset["DMT"] = "dmt";
    UnderlyingAsset["DNA"] = "dna";
    UnderlyingAsset["DNT"] = "dnt";
    UnderlyingAsset["DODO"] = "dodo";
    UnderlyingAsset["DOG"] = "dog";
    UnderlyingAsset["DOGE"] = "doge";
    UnderlyingAsset["DOGEOS"] = "dogeos";
    UnderlyingAsset["DOGEBEAR"] = "dogebear";
    UnderlyingAsset["DOGEBEAR2021"] = "dogebear2021";
    UnderlyingAsset["DOGEBULL"] = "dogebull";
    UnderlyingAsset["DOMI"] = "domi";
    UnderlyingAsset["DOOM"] = "doom";
    UnderlyingAsset["DOOMSHIT"] = "doomshit";
    UnderlyingAsset["DOSE"] = "dose";
    UnderlyingAsset["DOTK"] = "dotk";
    UnderlyingAsset["DPAY"] = "dpay";
    UnderlyingAsset["DPI"] = "dpi";
    UnderlyingAsset["DPX"] = "dpx";
    UnderlyingAsset["DPY"] = "dpy";
    UnderlyingAsset["DRAM"] = "dram";
    UnderlyingAsset["DRGNBEAR"] = "drgnbear";
    UnderlyingAsset["DRGNBULL"] = "drgnbull";
    UnderlyingAsset["DRPU"] = "drpu";
    UnderlyingAsset["DRV"] = "drv";
    UnderlyingAsset["DUC"] = "duc";
    UnderlyingAsset["DUCK"] = "duck";
    UnderlyingAsset["DUSD"] = "dusd";
    UnderlyingAsset["DUSK"] = "dusk";
    UnderlyingAsset["DUST"] = "dust";
    UnderlyingAsset["DX1U"] = "dx1u";
    UnderlyingAsset["DXGT"] = "dxgt";
    UnderlyingAsset["DXO"] = "dxo";
    UnderlyingAsset["DXPT"] = "dxpt";
    UnderlyingAsset["DXST"] = "dxst";
    UnderlyingAsset["DYDX"] = "dydx";
    UnderlyingAsset["DYN"] = "dyn";
    UnderlyingAsset["EASY"] = "easy";
    UnderlyingAsset["EBTCQ"] = "ebtcq";
    UnderlyingAsset["ECHT"] = "echt";
    UnderlyingAsset["eth:eco"] = "eth:eco";
    UnderlyingAsset["ECOX"] = "ecox";
    UnderlyingAsset["sol:eusx"] = "sol:eusx";
    UnderlyingAsset["EDEN"] = "eden";
    UnderlyingAsset["EDISON"] = "edison";
    UnderlyingAsset["EDLC"] = "edlc";
    UnderlyingAsset["EDO"] = "edo";
    UnderlyingAsset["ELON"] = "elon";
    UnderlyingAsset["EMB"] = "emb";
    UnderlyingAsset["EDN"] = "edn";
    UnderlyingAsset["EDR"] = "edr";
    UnderlyingAsset["EFI"] = "efi";
    UnderlyingAsset["EGL"] = "egl";
    UnderlyingAsset["EGLD"] = "egld";
    UnderlyingAsset["EGOLD"] = "egold";
    UnderlyingAsset["EIGEN"] = "eigen";
    UnderlyingAsset["ELF"] = "elf";
    UnderlyingAsset["ELU"] = "elu";
    UnderlyingAsset["EMAID"] = "emaid";
    UnderlyingAsset["EMX"] = "emx";
    UnderlyingAsset["ENA"] = "ena";
    UnderlyingAsset["ENG"] = "eng";
    UnderlyingAsset["ENJ"] = "enj";
    UnderlyingAsset["ENS"] = "ens";
    UnderlyingAsset["EON"] = "eon";
    UnderlyingAsset["EOP"] = "eop";
    UnderlyingAsset["EOSBEAR"] = "eosbear";
    UnderlyingAsset["EOSBULL"] = "eosbull";
    UnderlyingAsset["EOSDOOM"] = "eosdoom";
    UnderlyingAsset["EOSHEDGE"] = "eoshedge";
    UnderlyingAsset["EOSMOON"] = "eosmoon";
    UnderlyingAsset["EQO"] = "eqo";
    UnderlyingAsset["ESE"] = "ese";
    UnderlyingAsset["ETA"] = "eta";
    UnderlyingAsset["ETHBULL"] = "ethbull";
    UnderlyingAsset["ETCBEAR"] = "etcbear";
    UnderlyingAsset["ETCBULL"] = "etcbull";
    UnderlyingAsset["ETCDOOM"] = "etcdoom";
    UnderlyingAsset["ETCHEDOOM"] = "etchedoom";
    UnderlyingAsset["ETCMOON"] = "etcmoon";
    UnderlyingAsset["ETHBEAR"] = "ethbear";
    UnderlyingAsset["ETHDOOM"] = "ethdoom";
    UnderlyingAsset["ETHFI"] = "ethfi";
    UnderlyingAsset["eth:block"] = "eth:block";
    UnderlyingAsset["eth:bito"] = "eth:bito";
    UnderlyingAsset["ETH:ECASH"] = "eth:ecash";
    UnderlyingAsset["ETH:OORT"] = "eth:oort";
    UnderlyingAsset["eth:prism"] = "eth:prism";
    UnderlyingAsset["eth:ultra"] = "eth:ultra";
    UnderlyingAsset["eth:yprism"] = "eth:yprism";
    UnderlyingAsset["eth:nvylds"] = "eth:nvylds";
    UnderlyingAsset["eth:nvheloc"] = "eth:nvheloc";
    UnderlyingAsset["eth:nil"] = "eth:nil";
    UnderlyingAsset["eth:dragonx"] = "eth:dragonx";
    UnderlyingAsset["ETHHEDGE"] = "ethhedge";
    UnderlyingAsset["ETHMOON"] = "ethmoon";
    UnderlyingAsset["ETHOPT"] = "ethopt";
    UnderlyingAsset["ETHOS"] = "ethos";
    UnderlyingAsset["ETHTON"] = "ethton";
    UnderlyingAsset["ETHX"] = "ethx";
    UnderlyingAsset["ETV"] = "etv";
    UnderlyingAsset["ETX"] = "etx";
    UnderlyingAsset["EUL"] = "eul";
    UnderlyingAsset["EURE"] = "eure";
    UnderlyingAsset["EURL"] = "eurl";
    UnderlyingAsset["EUROE"] = "euroe";
    UnderlyingAsset["EUROP"] = "europ";
    UnderlyingAsset["EURS"] = "eurs";
    UnderlyingAsset["EURST"] = "eurst";
    UnderlyingAsset["EURT"] = "eurt";
    UnderlyingAsset["EURX"] = "eurx";
    UnderlyingAsset["EUX"] = "eux";
    UnderlyingAsset["EVER"] = "ever";
    UnderlyingAsset["EVERY"] = "every";
    UnderlyingAsset["EVRY"] = "evry";
    UnderlyingAsset["EVX"] = "evx";
    UnderlyingAsset["EXCHBEAR"] = "exchbear";
    UnderlyingAsset["EXCHBULL"] = "exchbull";
    UnderlyingAsset["EXCHDOOM"] = "exchdoom";
    UnderlyingAsset["EXCHHEDGE"] = "exchhedge";
    UnderlyingAsset["EXCHMOON"] = "exchmoon";
    UnderlyingAsset["EXE"] = "exe";
    UnderlyingAsset["FANT"] = "fant";
    UnderlyingAsset["FARM"] = "farm";
    UnderlyingAsset["FEI"] = "fei";
    UnderlyingAsset["FET"] = "fet";
    UnderlyingAsset["FET1"] = "fet1";
    UnderlyingAsset["FDT"] = "fdt";
    UnderlyingAsset["FDUSD"] = "fdusd";
    UnderlyingAsset["FF1"] = "ff1";
    UnderlyingAsset["FF6000"] = "ff6000";
    UnderlyingAsset["FFT"] = "fft";
    UnderlyingAsset["FIDA"] = "fida";
    UnderlyingAsset["FIDU"] = "fidu";
    UnderlyingAsset["FIN"] = "fin";
    UnderlyingAsset["FIRE"] = "fire";
    UnderlyingAsset["FIRSTBLOOD"] = "firstblood";
    UnderlyingAsset["FIS"] = "fis";
    UnderlyingAsset["FIXED"] = "fixed";
    UnderlyingAsset["FLIP"] = "flip";
    UnderlyingAsset["FLOKI"] = "floki";
    UnderlyingAsset["FLUX"] = "flux";
    UnderlyingAsset["FLY"] = "fly";
    UnderlyingAsset["FMF"] = "fmf";
    UnderlyingAsset["FOLD"] = "fold";
    UnderlyingAsset["FOR"] = "for";
    UnderlyingAsset["sol:ford"] = "sol:ford";
    UnderlyingAsset["FOREX"] = "forex";
    UnderlyingAsset["FORT"] = "fort";
    UnderlyingAsset["FORTH"] = "forth";
    UnderlyingAsset["FOX"] = "fox";
    UnderlyingAsset["FPIS"] = "fpis";
    UnderlyingAsset["FRAX"] = "frax";
    UnderlyingAsset["FRONT"] = "front";
    UnderlyingAsset["FT"] = "ft";
    UnderlyingAsset["FTM"] = "ftm";
    UnderlyingAsset["FTT"] = "ftt";
    UnderlyingAsset["FTT20"] = "ftt20";
    UnderlyingAsset["FTX2"] = "ftx2";
    UnderlyingAsset["FUCKFTX"] = "fuckftx";
    UnderlyingAsset["FUN"] = "fun";
    UnderlyingAsset["FWB"] = "fwb";
    UnderlyingAsset["FX"] = "fx";
    UnderlyingAsset["FXRT"] = "fxrt";
    UnderlyingAsset["FXS"] = "fxs";
    UnderlyingAsset["G"] = "g";
    UnderlyingAsset["GAL"] = "gal";
    UnderlyingAsset["GALA"] = "gala";
    UnderlyingAsset["GALAV2"] = "galav2";
    UnderlyingAsset["GAME.COM"] = "game.com";
    UnderlyingAsset["GAMMA"] = "gamma";
    UnderlyingAsset["sol:gari"] = "sol:gari";
    UnderlyingAsset["sol:usd1"] = "sol:usd1";
    UnderlyingAsset["sol:usdm1"] = "sol:usdm1";
    UnderlyingAsset["tsol:slnd"] = "tsol:slnd";
    UnderlyingAsset["tsol:orca"] = "tsol:orca";
    UnderlyingAsset["tsol:usdc"] = "tsol:usdc";
    UnderlyingAsset["tsol:ray"] = "tsol:ray";
    UnderlyingAsset["tsol:gmt"] = "tsol:gmt";
    UnderlyingAsset["tsol:usdt"] = "tsol:usdt";
    UnderlyingAsset["tsol:srm"] = "tsol:srm";
    UnderlyingAsset["tsol:wsol"] = "tsol:wsol";
    UnderlyingAsset["tsol:gari"] = "tsol:gari";
    UnderlyingAsset["tsol:t22mint"] = "tsol:t22mint";
    UnderlyingAsset["tsol:t1test"] = "tsol:t1test";
    UnderlyingAsset["GAS"] = "gas";
    UnderlyingAsset["GATE"] = "gate";
    UnderlyingAsset["GBPT"] = "gbpt";
    UnderlyingAsset["GBPX"] = "gbpx";
    UnderlyingAsset["GDT"] = "gdt";
    UnderlyingAsset["GEAR"] = "gear";
    UnderlyingAsset["GEC"] = "gec";
    UnderlyingAsset["GEL"] = "gel";
    UnderlyingAsset["GEN"] = "gen";
    UnderlyingAsset["GENE"] = "gene";
    UnderlyingAsset["GENIE"] = "genie";
    UnderlyingAsset["GF"] = "gf";
    UnderlyingAsset["GFI"] = "gfi";
    UnderlyingAsset["GHST"] = "ghst";
    UnderlyingAsset["GHUB"] = "ghub";
    UnderlyingAsset["GIGDROP"] = "gigdrop";
    UnderlyingAsset["GIV"] = "giv";
    UnderlyingAsset["GLDX"] = "gldx";
    UnderlyingAsset["GLM"] = "glm";
    UnderlyingAsset["sol:glxy"] = "sol:glxy";
    UnderlyingAsset["GMT"] = "gmt";
    UnderlyingAsset["sol:gmt"] = "sol:gmt";
    UnderlyingAsset["GNO"] = "gno";
    UnderlyingAsset["GNT"] = "gnt";
    UnderlyingAsset["sol:goat"] = "sol:goat";
    UnderlyingAsset["GODS"] = "gods";
    UnderlyingAsset["GOHM"] = "gohm";
    UnderlyingAsset["GOG"] = "gog";
    UnderlyingAsset["GOLD"] = "gold";
    UnderlyingAsset["GOM"] = "gom";
    UnderlyingAsset["GOMINING"] = "gomining";
    UnderlyingAsset["GOT"] = "got";
    UnderlyingAsset["GRID"] = "grid";
    UnderlyingAsset["GRT"] = "grt";
    UnderlyingAsset["GST"] = "gst";
    UnderlyingAsset["GT"] = "gt";
    UnderlyingAsset["GTAAVE18DP"] = "gtaave18dp";
    UnderlyingAsset["GTBAT18DP"] = "gtbat18dp";
    UnderlyingAsset["GTCOMP18DP"] = "gtcomp18dp";
    UnderlyingAsset["GTGRT18DP"] = "gtgrt18dp";
    UnderlyingAsset["GTLINK18DP"] = "gtlink18dp";
    UnderlyingAsset["GTMKR18DP"] = "gtmkr18dp";
    UnderlyingAsset["GTSNX18DP"] = "gtsnx18dp";
    UnderlyingAsset["GTUNI18DP"] = "gtuni18dp";
    UnderlyingAsset["GTUSDT6DP"] = "gtusdt6dp";
    UnderlyingAsset["GTYFI18DP"] = "gtyfi18dp";
    UnderlyingAsset["GTWBTC8DP"] = "gtwbtc8dp";
    UnderlyingAsset["GTO"] = "gto";
    UnderlyingAsset["GTERC2DP"] = "gterc2dp";
    UnderlyingAsset["GTERC6DP"] = "gterc6dp";
    UnderlyingAsset["GTERC18DP"] = "gterc18dp";
    UnderlyingAsset["GUSD"] = "gusd";
    UnderlyingAsset["GUSDT"] = "gusdt";
    UnderlyingAsset["GXC"] = "gxc";
    UnderlyingAsset["GXT"] = "gxt";
    UnderlyingAsset["GYEN"] = "gyen";
    UnderlyingAsset["HBB"] = "hbb";
    UnderlyingAsset["HBTC"] = "hbtc";
    UnderlyingAsset["HCN"] = "hcn";
    UnderlyingAsset["HDO"] = "hdo";
    UnderlyingAsset["HEDG"] = "hedg";
    UnderlyingAsset["HEDGE"] = "hedge";
    UnderlyingAsset["HEDGESHIT"] = "hedgeshit";
    UnderlyingAsset["HEX"] = "hex";
    UnderlyingAsset["HFT"] = "hft";
    UnderlyingAsset["HGET"] = "hget";
    UnderlyingAsset["HIGH"] = "high";
    UnderlyingAsset["HIFI"] = "hifi";
    UnderlyingAsset["HIT"] = "hit";
    UnderlyingAsset["HKDX"] = "hkdx";
    UnderlyingAsset["HLC"] = "hlc";
    UnderlyingAsset["HMT"] = "hmt";
    UnderlyingAsset["sol:hnt"] = "sol:hnt";
    UnderlyingAsset["HOLD"] = "hold";
    UnderlyingAsset["HOLY"] = "holy";
    UnderlyingAsset["HOP"] = "hop";
    UnderlyingAsset["HOT"] = "hot";
    UnderlyingAsset["HPO"] = "hpo";
    UnderlyingAsset["HQG"] = "hqg";
    UnderlyingAsset["HQT"] = "hqt";
    UnderlyingAsset["HST"] = "hst";
    UnderlyingAsset["HT"] = "ht";
    UnderlyingAsset["HTBEAR"] = "htbear";
    UnderlyingAsset["HTBULL"] = "htbull";
    UnderlyingAsset["HTDOOM"] = "htdoom";
    UnderlyingAsset["hteth:bgerchv2"] = "hteth:bgerchv2";
    UnderlyingAsset["hteth:aut"] = "hteth:aut";
    UnderlyingAsset["HTHEDGE"] = "hthedge";
    UnderlyingAsset["HTMOON"] = "htmoon";
    UnderlyingAsset["HUM"] = "hum";
    UnderlyingAsset["HUMV2"] = "humv2";
    UnderlyingAsset["HUSD"] = "husd";
    UnderlyingAsset["HXRO"] = "hxro";
    UnderlyingAsset["HYB"] = "hyb";
    UnderlyingAsset["HYDRO"] = "hydro";
    UnderlyingAsset["HYDROPROTOCOL"] = "hydroprotocol";
    UnderlyingAsset["I8"] = "i8";
    UnderlyingAsset["IBEUR"] = "ibeur";
    UnderlyingAsset["IBOX"] = "ibox";
    UnderlyingAsset["IBVOL"] = "ibvol";
    UnderlyingAsset["ICETH"] = "iceth";
    UnderlyingAsset["ID"] = "id";
    UnderlyingAsset["IDEX"] = "idex";
    UnderlyingAsset["IDRC"] = "idrc";
    UnderlyingAsset["IDRT"] = "idrt";
    UnderlyingAsset["ILV"] = "ilv";
    UnderlyingAsset["IMX"] = "imx";
    UnderlyingAsset["IMXV2"] = "imxv2";
    UnderlyingAsset["INCX"] = "incx";
    UnderlyingAsset["IND"] = "ind";
    UnderlyingAsset["INDEX"] = "index";
    UnderlyingAsset["INDI"] = "indi";
    UnderlyingAsset["INF"] = "inf";
    UnderlyingAsset["INJ"] = "inj";
    UnderlyingAsset["INJV2"] = "injv2";
    UnderlyingAsset["INST"] = "inst";
    UnderlyingAsset["INSUR"] = "insur";
    UnderlyingAsset["INV"] = "inv";
    UnderlyingAsset["INX"] = "inx";
    UnderlyingAsset["IOST"] = "iost";
    UnderlyingAsset["IOTX"] = "iotx";
    UnderlyingAsset["IP3"] = "ip3";
    UnderlyingAsset["ISF"] = "isf";
    UnderlyingAsset["ISR"] = "isr";
    UnderlyingAsset["IVO"] = "ivo";
    UnderlyingAsset["IVY"] = "ivy";
    UnderlyingAsset["JASMY"] = "jasmy";
    UnderlyingAsset["JBC"] = "jbc";
    UnderlyingAsset["JCR"] = "jcr";
    UnderlyingAsset["JCG"] = "jcg";
    UnderlyingAsset["sol:jet"] = "sol:jet";
    UnderlyingAsset["JFIN"] = "jfin";
    UnderlyingAsset["JPYX"] = "jpyx";
    UnderlyingAsset["JSOL"] = "jsol";
    UnderlyingAsset["KARATE"] = "karate";
    UnderlyingAsset["KARMA"] = "karma";
    UnderlyingAsset["KAS"] = "kas";
    UnderlyingAsset["KCASH"] = "kcash";
    UnderlyingAsset["KCS"] = "kcs";
    UnderlyingAsset["KEEP"] = "keep";
    UnderlyingAsset["KEY"] = "key";
    UnderlyingAsset["KILL0"] = "kill0";
    UnderlyingAsset["KIN"] = "kin";
    UnderlyingAsset["sol:kin"] = "sol:kin";
    UnderlyingAsset["KINE"] = "kine";
    UnderlyingAsset["KING"] = "king";
    UnderlyingAsset["KINTO"] = "kinto";
    UnderlyingAsset["KIRO"] = "kiro";
    UnderlyingAsset["KISHUI"] = "kishui";
    UnderlyingAsset["KITTY"] = "kitty";
    UnderlyingAsset["KNC"] = "knc";
    UnderlyingAsset["KNC2"] = "knc2";
    UnderlyingAsset["KOIN"] = "koin";
    UnderlyingAsset["KOL"] = "kol";
    UnderlyingAsset["KOZ"] = "koz";
    UnderlyingAsset["KP3R"] = "kp3r";
    UnderlyingAsset["KRO"] = "kro";
    UnderlyingAsset["KROM"] = "krom";
    UnderlyingAsset["KTRC"] = "ktrc";
    UnderlyingAsset["KZE"] = "kze";
    UnderlyingAsset["L3"] = "l3";
    UnderlyingAsset["L3USD"] = "l3usd";
    UnderlyingAsset["LA"] = "la";
    UnderlyingAsset["LADYS"] = "ladys";
    UnderlyingAsset["LAYER"] = "layer";
    UnderlyingAsset["LAYERZERO"] = "layerzero";
    UnderlyingAsset["LBA"] = "lba";
    UnderlyingAsset["LCX"] = "lcx";
    UnderlyingAsset["LDO"] = "ldo";
    UnderlyingAsset["LEND"] = "lend";
    UnderlyingAsset["LEO"] = "leo";
    UnderlyingAsset["LEOBEAR"] = "leobear";
    UnderlyingAsset["LEOBULL"] = "leobull";
    UnderlyingAsset["LEODOOM"] = "leodoom";
    UnderlyingAsset["LEOHEDGE"] = "leohedge";
    UnderlyingAsset["LEOMOON"] = "leomoon";
    UnderlyingAsset["LEV"] = "lev";
    UnderlyingAsset["LEVER"] = "lever";
    UnderlyingAsset["LGO"] = "lgo";
    UnderlyingAsset["LIEN"] = "lien";
    UnderlyingAsset["LIF3"] = "lif3";
    UnderlyingAsset["LIKE"] = "like";
    UnderlyingAsset["LINA"] = "lina";
    UnderlyingAsset["LINK"] = "link";
    UnderlyingAsset["LINKBEAR"] = "linkbear";
    UnderlyingAsset["LINKBULL"] = "linkbull";
    UnderlyingAsset["LION"] = "lion";
    UnderlyingAsset["LIT"] = "lit";
    UnderlyingAsset["LITH"] = "lith";
    UnderlyingAsset["LITv2"] = "litv2";
    UnderlyingAsset["LKR"] = "lkr";
    UnderlyingAsset["LMWR"] = "lmwr";
    UnderlyingAsset["LNC"] = "lnc";
    UnderlyingAsset["LOKA"] = "loka";
    UnderlyingAsset["LOOKS"] = "looks";
    UnderlyingAsset["LOOM"] = "loom";
    UnderlyingAsset["LOOM1"] = "loom1";
    UnderlyingAsset["LOVE"] = "love";
    UnderlyingAsset["LOVELY"] = "lovely";
    UnderlyingAsset["LOWB"] = "lowb";
    UnderlyingAsset["LPT"] = "lpt";
    UnderlyingAsset["LQID"] = "lqid";
    UnderlyingAsset["LQTY"] = "lqty";
    UnderlyingAsset["LRC"] = "lrc";
    UnderlyingAsset["LRCV2"] = "lrcv2";
    UnderlyingAsset["LSETH"] = "lseth";
    UnderlyingAsset["LSK"] = "lsk";
    UnderlyingAsset["LTCBEAR"] = "ltcbear";
    UnderlyingAsset["LTCBULL"] = "ltcbull";
    UnderlyingAsset["LTCDOOM"] = "ltcdoom";
    UnderlyingAsset["LTCHEDGE"] = "ltchedge";
    UnderlyingAsset["LTCMOON"] = "ltcmoon";
    UnderlyingAsset["LTO"] = "lto";
    UnderlyingAsset["LUA"] = "lua";
    UnderlyingAsset["LUNA"] = "luna";
    UnderlyingAsset["LUNAWORMHOLE"] = "lunawormhole";
    UnderlyingAsset["LYN"] = "lyn";
    UnderlyingAsset["LYXE"] = "lyxe";
    UnderlyingAsset["MAGIC"] = "magic";
    UnderlyingAsset["MANA"] = "mana";
    UnderlyingAsset["MAPS"] = "maps";
    UnderlyingAsset["MASA"] = "masa";
    UnderlyingAsset["MASK"] = "mask";
    UnderlyingAsset["MATH"] = "math";
    UnderlyingAsset["MATIC"] = "matic";
    UnderlyingAsset["MATICBEAR"] = "maticbear";
    UnderlyingAsset["MATICBEAR2021"] = "maticbear2021";
    UnderlyingAsset["MATICBULL"] = "maticbull";
    UnderlyingAsset["MATTER"] = "matter";
    UnderlyingAsset["MAV"] = "mav";
    UnderlyingAsset["MBS"] = "mbs";
    UnderlyingAsset["MCAU"] = "mcau";
    UnderlyingAsset["MCB"] = "mcb";
    UnderlyingAsset["MCDAI"] = "mcdai";
    UnderlyingAsset["MCO"] = "mco";
    UnderlyingAsset["MCO2"] = "mco2";
    UnderlyingAsset["MCS"] = "mcs";
    UnderlyingAsset["MCX"] = "mcx";
    UnderlyingAsset["MDFC"] = "mdfc";
    UnderlyingAsset["MDT"] = "mdt";
    UnderlyingAsset["MDX"] = "mdx";
    UnderlyingAsset["MEAN"] = "mean";
    UnderlyingAsset["MEDIA"] = "media";
    UnderlyingAsset["MEDIAv2"] = "mediav2";
    UnderlyingAsset["MEDX"] = "medx";
    UnderlyingAsset["MEME"] = "meme";
    UnderlyingAsset["MEOW"] = "meow";
    UnderlyingAsset["MER"] = "mer";
    UnderlyingAsset["MET"] = "met";
    UnderlyingAsset["META"] = "meta";
    UnderlyingAsset["METIS"] = "metis";
    UnderlyingAsset["MEW"] = "mew";
    UnderlyingAsset["MFG"] = "mfg";
    UnderlyingAsset["MFPH"] = "mfph";
    UnderlyingAsset["MFT"] = "mft";
    UnderlyingAsset["MIDBEAR"] = "midbear";
    UnderlyingAsset["MIDBULL"] = "midbull";
    UnderlyingAsset["MIDDOOM"] = "middoom";
    UnderlyingAsset["MIDHEDGE"] = "midhedge";
    UnderlyingAsset["MIDMOON"] = "midmoon";
    UnderlyingAsset["MILKV2"] = "milkv2";
    UnderlyingAsset["MIM"] = "mim";
    UnderlyingAsset["MIR"] = "mir";
    UnderlyingAsset["MITH"] = "mith";
    UnderlyingAsset["MIX"] = "mix";
    UnderlyingAsset["MIZN"] = "mizn";
    UnderlyingAsset["MKR"] = "mkr";
    UnderlyingAsset["MLN"] = "mln";
    UnderlyingAsset["MNS"] = "mns";
    UnderlyingAsset["MNT"] = "mnt";
    UnderlyingAsset["MNDE"] = "mnde";
    UnderlyingAsset["sol:mnde"] = "sol:mnde";
    UnderlyingAsset["MOC"] = "moc";
    UnderlyingAsset["MOCA"] = "moca";
    UnderlyingAsset["MOCHI"] = "mochi";
    UnderlyingAsset["MOF"] = "mof";
    UnderlyingAsset["MOG"] = "mog";
    UnderlyingAsset["MOH"] = "moh";
    UnderlyingAsset["MOON"] = "moon";
    UnderlyingAsset["MOONSHIT"] = "moonshit";
    UnderlyingAsset["MOTHER"] = "mother";
    UnderlyingAsset["MNGO"] = "mngo";
    UnderlyingAsset["MPAY"] = "mpay";
    UnderlyingAsset["MPL"] = "mpl";
    UnderlyingAsset["sol:mplx"] = "sol:mplx";
    UnderlyingAsset["MRTWEET"] = "mrtweet";
    UnderlyingAsset["MSN"] = "msn";
    UnderlyingAsset["MSOL"] = "msol";
    UnderlyingAsset["MTA"] = "mta";
    UnderlyingAsset["MTCN"] = "mtcn";
    UnderlyingAsset["MTH"] = "mth";
    UnderlyingAsset["MTL"] = "mtl";
    UnderlyingAsset["MTV"] = "mtv";
    UnderlyingAsset["MUSD"] = "musd";
    UnderlyingAsset["MVL"] = "mvl";
    UnderlyingAsset["MVI"] = "mvi";
    UnderlyingAsset["MWT"] = "mwt";
    UnderlyingAsset["MYRC"] = "myrc";
    UnderlyingAsset["sol:myrc"] = "sol:myrc";
    UnderlyingAsset["MYTH"] = "myth";
    UnderlyingAsset["NAAI"] = "naai";
    UnderlyingAsset["NAS"] = "nas";
    UnderlyingAsset["NCT"] = "nct";
    UnderlyingAsset["NDX"] = "ndx";
    UnderlyingAsset["NEAR-ERC20"] = "near-erc20";
    UnderlyingAsset["NEU"] = "neu";
    UnderlyingAsset["NEWO"] = "newo";
    UnderlyingAsset["NEXO"] = "nexo";
    UnderlyingAsset["NFCWIN-SB-2021"] = "nfcwin-sb-2021";
    UnderlyingAsset["NFTFI"] = "nftfi";
    UnderlyingAsset["NFTX"] = "nftx";
    UnderlyingAsset["NGNT"] = "ngnt";
    UnderlyingAsset["NIAX"] = "niax";
    UnderlyingAsset["NKN"] = "nkn";
    UnderlyingAsset["NMR"] = "nmr";
    UnderlyingAsset["NOSANA"] = "nosana";
    UnderlyingAsset["NOTE"] = "note";
    UnderlyingAsset["NOVA"] = "nova";
    UnderlyingAsset["NPT"] = "npt";
    UnderlyingAsset["NPXS"] = "npxs";
    UnderlyingAsset["NS2DRP"] = "ns2drp";
    UnderlyingAsset["NU"] = "nu";
    UnderlyingAsset["NULS"] = "nuls";
    UnderlyingAsset["NUTS"] = "nuts";
    UnderlyingAsset["NXM"] = "nxm";
    UnderlyingAsset["NYM"] = "nym";
    UnderlyingAsset["NZDX"] = "nzdx";
    UnderlyingAsset["OAX"] = "oax";
    UnderlyingAsset["OCEAN"] = "ocean";
    UnderlyingAsset["OCEANV2"] = "oceanv2";
    UnderlyingAsset["OCTAV"] = "octav";
    UnderlyingAsset["OGN"] = "ogn";
    UnderlyingAsset["OGV"] = "ogv";
    UnderlyingAsset["OKB"] = "okb";
    UnderlyingAsset["OKBBEAR"] = "okbbear";
    UnderlyingAsset["OKBBULL"] = "okbbull";
    UnderlyingAsset["OKBDOOM"] = "okbdoom";
    UnderlyingAsset["OKBHEDGE"] = "okbhedge";
    UnderlyingAsset["OKBMOON"] = "okbmoon";
    UnderlyingAsset["OM"] = "om";
    UnderlyingAsset["OMOLD"] = "omold";
    UnderlyingAsset["OMG"] = "omg";
    UnderlyingAsset["OMNI"] = "omni";
    UnderlyingAsset["OMNIA"] = "omnia";
    UnderlyingAsset["ONDO"] = "ondo";
    UnderlyingAsset["ONL"] = "onl";
    UnderlyingAsset["ONT"] = "ont";
    UnderlyingAsset["OOKI"] = "ooki";
    UnderlyingAsset["OP"] = "op";
    UnderlyingAsset["OPIUM"] = "opium";
    UnderlyingAsset["OPT"] = "opt";
    UnderlyingAsset["ORAI"] = "orai";
    UnderlyingAsset["ORBS"] = "orbs";
    UnderlyingAsset["ORC"] = "orc";
    UnderlyingAsset["ORN"] = "orn";
    UnderlyingAsset["sol:orca"] = "sol:orca";
    UnderlyingAsset["OS"] = "os";
    UnderlyingAsset["OSETH"] = "oseth";
    UnderlyingAsset["OUSD"] = "ousd";
    UnderlyingAsset["OUSG"] = "ousg";
    UnderlyingAsset["OWN"] = "own";
    UnderlyingAsset["OXT"] = "oxt";
    UnderlyingAsset["OXY"] = "oxy";
    UnderlyingAsset["OHM"] = "ohm";
    UnderlyingAsset["PACT"] = "pact";
    UnderlyingAsset["PAI"] = "pai";
    UnderlyingAsset["PAR"] = "par";
    UnderlyingAsset["PASS"] = "pass";
    UnderlyingAsset["PAU"] = "pau";
    UnderlyingAsset["PAX"] = "pax";
    UnderlyingAsset["PAXG"] = "paxg";
    UnderlyingAsset["PAXGBEAR"] = "paxgbear";
    UnderlyingAsset["PAXGBULL"] = "paxgbull";
    UnderlyingAsset["PAY"] = "pay";
    UnderlyingAsset["PBCH"] = "pbch";
    UnderlyingAsset["PBTC"] = "pbtc";
    UnderlyingAsset["PDA"] = "PDA";
    UnderlyingAsset["PDATA"] = "pdata";
    UnderlyingAsset["PDI"] = "pdi";
    UnderlyingAsset["PEAQ"] = "peaq";
    UnderlyingAsset["PEBBLE"] = "pebble";
    UnderlyingAsset["PEG"] = "peg";
    UnderlyingAsset["PENDLE"] = "pendle";
    UnderlyingAsset["PEOPLE"] = "people";
    UnderlyingAsset["PEPE"] = "pepe";
    UnderlyingAsset["PERL"] = "perl";
    UnderlyingAsset["PERP"] = "perp";
    UnderlyingAsset["PETH"] = "peth";
    UnderlyingAsset["PHA"] = "pha";
    UnderlyingAsset["PHNX"] = "phnx";
    UnderlyingAsset["PICK"] = "pick";
    UnderlyingAsset["PICKLE"] = "pickle";
    UnderlyingAsset["PIE"] = "pie";
    UnderlyingAsset["PINE"] = "pine";
    UnderlyingAsset["PIRATE"] = "pirate";
    UnderlyingAsset["PLAY"] = "play";
    UnderlyingAsset["PIXEL"] = "pixel";
    UnderlyingAsset["PLC"] = "plc";
    UnderlyingAsset["PFCT"] = "pfct";
    UnderlyingAsset["PLANET"] = "planet";
    UnderlyingAsset["PLNX"] = "plnx";
    UnderlyingAsset["PLX"] = "plx";
    UnderlyingAsset["PMA"] = "pma";
    UnderlyingAsset["PNT"] = "pnt";
    UnderlyingAsset["POL"] = "pol";
    UnderlyingAsset["POLIS"] = "polis";
    UnderlyingAsset["POLY"] = "poly";
    UnderlyingAsset["POLYX"] = "polyx";
    UnderlyingAsset["POLS"] = "pols";
    UnderlyingAsset["POND"] = "pond";
    UnderlyingAsset["PONYS"] = "ponys";
    UnderlyingAsset["PORT"] = "port";
    UnderlyingAsset["POWR"] = "powr";
    UnderlyingAsset["PPT"] = "ppt";
    UnderlyingAsset["PRDX"] = "prdx";
    UnderlyingAsset["PRINTS"] = "prints";
    UnderlyingAsset["PRISM"] = "prism";
    UnderlyingAsset["PRO"] = "pro";
    UnderlyingAsset["PROM"] = "prom";
    UnderlyingAsset["PROS"] = "pros";
    UnderlyingAsset["PRT"] = "prt";
    UnderlyingAsset["PRTS"] = "prts";
    UnderlyingAsset["PSOL"] = "psol";
    UnderlyingAsset["PSP"] = "psp";
    UnderlyingAsset["PSTAKE"] = "pstake";
    UnderlyingAsset["PSY"] = "psy";
    UnderlyingAsset["PTU"] = "ptu";
    UnderlyingAsset["PUNDIX"] = "pundix";
    UnderlyingAsset["sol:pump"] = "sol:pump";
    UnderlyingAsset["PUSD"] = "pusd";
    UnderlyingAsset["PUSH"] = "push";
    UnderlyingAsset["PV01"] = "pv01";
    UnderlyingAsset["PXP"] = "pxp";
    UnderlyingAsset["PYR"] = "pyr";
    UnderlyingAsset["PYUSD"] = "pyusd";
    UnderlyingAsset["QASH"] = "qash";
    UnderlyingAsset["QCAD"] = "qcad";
    UnderlyingAsset["sol:qcad"] = "sol:qcad";
    UnderlyingAsset["QOM"] = "qom";
    UnderlyingAsset["QUICK"] = "quick";
    UnderlyingAsset["QDT"] = "qdt";
    UnderlyingAsset["QKC"] = "qkc";
    UnderlyingAsset["QLINDO"] = "qlindo";
    UnderlyingAsset["QNT"] = "qnt";
    UnderlyingAsset["QRDO"] = "qrdo";
    UnderlyingAsset["QRL"] = "qrl";
    UnderlyingAsset["QSP"] = "qsp";
    UnderlyingAsset["QVT"] = "qvt";
    UnderlyingAsset["RAD"] = "rad";
    UnderlyingAsset["RADAR"] = "radar";
    UnderlyingAsset["RAIN"] = "rain";
    UnderlyingAsset["RALPH"] = "ralph";
    UnderlyingAsset["RAMP"] = "ramp";
    UnderlyingAsset["RARE"] = "rare";
    UnderlyingAsset["RARI"] = "rari";
    UnderlyingAsset["RAY"] = "ray";
    UnderlyingAsset["sol:ray"] = "sol:ray";
    UnderlyingAsset["RAZOR"] = "razor";
    UnderlyingAsset["RBANK"] = "rbank";
    UnderlyingAsset["RBN"] = "rbn";
    UnderlyingAsset["RBX"] = "rbx";
    UnderlyingAsset["RBY"] = "rby";
    UnderlyingAsset["RCOIN"] = "rcoin";
    UnderlyingAsset["RCT"] = "rct";
    UnderlyingAsset["RDN"] = "rdn";
    UnderlyingAsset["RDNT"] = "rdnt";
    UnderlyingAsset["REAL"] = "real";
    UnderlyingAsset["REB"] = "reb";
    UnderlyingAsset["REBL"] = "rebl";
    UnderlyingAsset["REEF"] = "reef";
    UnderlyingAsset["REF"] = "ref";
    UnderlyingAsset["REKTTOKEN"] = "rekttoken";
    UnderlyingAsset["REKTGAME"] = "rektgame";
    UnderlyingAsset["REN"] = "ren";
    UnderlyingAsset["RENBTC"] = "renbtc";
    UnderlyingAsset["RENDOGE"] = "rendoge";
    UnderlyingAsset["REP"] = "rep";
    UnderlyingAsset["REPV2"] = "repv2";
    UnderlyingAsset["REQ"] = "REQ";
    UnderlyingAsset["RETH-ROCKET"] = "reth-rocket";
    UnderlyingAsset["RETH-STAFI"] = "reth-stafi";
    UnderlyingAsset["RETH-H"] = "reth-h";
    UnderlyingAsset["RETH2"] = "reth2";
    UnderlyingAsset["REVV"] = "revv";
    UnderlyingAsset["REZ"] = "rez";
    UnderlyingAsset["RFOX"] = "rfox";
    UnderlyingAsset["RFR"] = "rfr";
    UnderlyingAsset["RFUEL"] = "rfuel";
    UnderlyingAsset["RGT"] = "rgt";
    UnderlyingAsset["RIF"] = "rif";
    UnderlyingAsset["RINGX"] = "ringx";
    UnderlyingAsset["RIO"] = "rio";
    UnderlyingAsset["RLC"] = "rlc";
    UnderlyingAsset["RLUSD"] = "rlusd";
    UnderlyingAsset["RLY"] = "rly";
    UnderlyingAsset["RN"] = "rn";
    UnderlyingAsset["RND"] = "rnd";
    UnderlyingAsset["RNDR"] = "rndr";
    UnderlyingAsset["RNDT"] = "rndt";
    UnderlyingAsset["ROOK"] = "rook";
    UnderlyingAsset["RON"] = "ron";
    UnderlyingAsset["RONC"] = "ronc";
    UnderlyingAsset["ROOBEE"] = "roobee";
    UnderlyingAsset["RPK"] = "rpk";
    UnderlyingAsset["RPL"] = "rpl";
    UnderlyingAsset["RSR"] = "rsr";
    UnderlyingAsset["RSWETH"] = "rsweth";
    UnderlyingAsset["RUBX"] = "rubx";
    UnderlyingAsset["RUEDATK"] = "ruedatk";
    UnderlyingAsset["RUN"] = "run";
    UnderlyingAsset["RUNE"] = "rune";
    UnderlyingAsset["RVR"] = "rvr";
    UnderlyingAsset["RYOSHI"] = "ryoshi";
    UnderlyingAsset["SAFE"] = "safe";
    UnderlyingAsset["SAIL"] = "sail";
    UnderlyingAsset["SAITABIT"] = "saitabit";
    UnderlyingAsset["SALT"] = "salt";
    UnderlyingAsset["SAND"] = "sand";
    UnderlyingAsset["SASHIMI"] = "sashimi";
    UnderlyingAsset["SAMO"] = "samo";
    UnderlyingAsset["SBC"] = "sbc";
    UnderlyingAsset["sol:sbc"] = "sol:sbc";
    UnderlyingAsset["sol:veur"] = "sol:veur";
    UnderlyingAsset["sol:vchf"] = "sol:vchf";
    UnderlyingAsset["sol:tbill"] = "sol:tbill";
    UnderlyingAsset["sol:usdg"] = "sol:usdg";
    UnderlyingAsset["sol:ausd"] = "sol:ausd";
    UnderlyingAsset["SBF"] = "sbf";
    UnderlyingAsset["SBR"] = "sbr";
    // Saber IOU Token (Liquidity Mining Rewards)
    UnderlyingAsset["SBRIOU"] = "sbriou";
    UnderlyingAsset["SCNSOL"] = "scnsol";
    UnderlyingAsset["SCOPE"] = "scope";
    UnderlyingAsset["SD"] = "sd";
    UnderlyingAsset["SDL"] = "sdl";
    UnderlyingAsset["SECO"] = "seco";
    UnderlyingAsset["SETH"] = "seth";
    UnderlyingAsset["SETH-H"] = "seth-h";
    UnderlyingAsset["SETH2"] = "seth2";
    UnderlyingAsset["SEWERCOIN"] = "sewercoin";
    UnderlyingAsset["SFI"] = "sfi";
    UnderlyingAsset["SGA"] = "sga";
    UnderlyingAsset["SGDX"] = "sgdx";
    UnderlyingAsset["SGR"] = "sgr";
    UnderlyingAsset["SGT"] = "sgt";
    UnderlyingAsset["SHDW"] = "shdw";
    UnderlyingAsset["SHEESH"] = "sheesh";
    UnderlyingAsset["SHIDO"] = "shido";
    UnderlyingAsset["SHK"] = "shk";
    UnderlyingAsset["SHOPX"] = "shopx";
    UnderlyingAsset["SHOW"] = "show";
    UnderlyingAsset["SHIB"] = "shib";
    UnderlyingAsset["SHR"] = "shr";
    UnderlyingAsset["SIH"] = "sih";
    UnderlyingAsset["SILV"] = "silv";
    UnderlyingAsset["SIPHER"] = "sipher";
    UnderlyingAsset["SIS"] = "sis";
    UnderlyingAsset["SKALE"] = "skale";
    UnderlyingAsset["SLAB"] = "slab";
    UnderlyingAsset["SLC"] = "slc";
    UnderlyingAsset["SLCL"] = "slcl";
    UnderlyingAsset["sol:slnd"] = "sol:slnd";
    UnderlyingAsset["SLOT"] = "slot";
    UnderlyingAsset["SLP"] = "slp";
    UnderlyingAsset["SLRS"] = "slrs";
    UnderlyingAsset["SLVX"] = "slvx";
    UnderlyingAsset["SMT"] = "smt";
    UnderlyingAsset["SNC"] = "snc";
    UnderlyingAsset["SNM"] = "snm";
    UnderlyingAsset["SNOV"] = "snov";
    UnderlyingAsset["SNT"] = "snt";
    UnderlyingAsset["SNX"] = "snx";
    UnderlyingAsset["SNY"] = "sny";
    UnderlyingAsset["SOC"] = "soc";
    UnderlyingAsset["SOHM"] = "sohm";
    UnderlyingAsset["SOMM"] = "somm";
    UnderlyingAsset["SOS"] = "sos";
    UnderlyingAsset["SPA"] = "spa";
    UnderlyingAsset["SPELL"] = "spell";
    UnderlyingAsset["SPF"] = "spf";
    UnderlyingAsset["SPO"] = "spo";
    UnderlyingAsset["SOLVE"] = "solve";
    UnderlyingAsset["SQUID2.0"] = "squid2.0";
    UnderlyingAsset["SRNT"] = "srnt";
    UnderlyingAsset["SRM"] = "srm";
    UnderlyingAsset["sol:srm"] = "sol:srm";
    UnderlyingAsset["sol:pipe"] = "sol:pipe";
    UnderlyingAsset["SSV"] = "ssv";
    UnderlyingAsset["STARS"] = "stars";
    UnderlyingAsset["STATE"] = "state";
    UnderlyingAsset["STBU"] = "stbu";
    UnderlyingAsset["STC"] = "stc";
    UnderlyingAsset["STCV2"] = "stcv2";
    UnderlyingAsset["STEP"] = "step";
    UnderlyingAsset["STETH"] = "steth";
    UnderlyingAsset["STG"] = "stg";
    UnderlyingAsset["STKAAVE"] = "stkaave";
    UnderlyingAsset["STMX"] = "stmx";
    UnderlyingAsset["STORE"] = "store";
    UnderlyingAsset["STORJ"] = "storj";
    UnderlyingAsset["STORM"] = "storm";
    UnderlyingAsset["STPT"] = "stpt";
    UnderlyingAsset["STRIKE"] = "strike";
    UnderlyingAsset["STRK"] = "strk";
    UnderlyingAsset["STRONG"] = "strong";
    UnderlyingAsset["STSOL"] = "stsol";
    UnderlyingAsset["STZEN"] = "stzen";
    UnderlyingAsset["SUI-ERC20"] = "sui-erc20";
    UnderlyingAsset["SUN"] = "sun";
    UnderlyingAsset["SUNNY"] = "sunny";
    UnderlyingAsset["SUPER"] = "super";
    UnderlyingAsset["SUPERPERIO"] = "superperio";
    UnderlyingAsset["SUSD"] = "susd";
    UnderlyingAsset["SUSDE"] = "susde";
    UnderlyingAsset["SUSHI"] = "sushi";
    UnderlyingAsset["SQUIG"] = "squig";
    UnderlyingAsset["SVT"] = "svt";
    UnderlyingAsset["SWAG"] = "swag";
    UnderlyingAsset["SWAP"] = "SWAP";
    UnderlyingAsset["SWEAT"] = "sweat";
    UnderlyingAsset["SWETH"] = "sweth";
    UnderlyingAsset["SWISE"] = "swise";
    UnderlyingAsset["SWITCH"] = "switch";
    UnderlyingAsset["SWRV"] = "swrv";
    UnderlyingAsset["SXP"] = "sxp";
    UnderlyingAsset["SYN"] = "syn";
    UnderlyingAsset["SYNCH"] = "synch";
    UnderlyingAsset["SYRUP"] = "syrup";
    UnderlyingAsset["SYNTH-SUSD"] = "synth-susd";
    UnderlyingAsset["TAO"] = "tao";
    UnderlyingAsset["THRESHOLD"] = "threshold";
    UnderlyingAsset["THEU"] = "theu";
    UnderlyingAsset["TAUD"] = "taud";
    UnderlyingAsset["TBILL"] = "tbill";
    UnderlyingAsset["TBTC1"] = "tbtc1";
    UnderlyingAsset["TBTC2"] = "tbtc2";
    UnderlyingAsset["TCAD"] = "tcad";
    UnderlyingAsset["TCO"] = "tco";
    UnderlyingAsset["TEIGEN"] = "teigen";
    UnderlyingAsset["TEINU"] = "teinu";
    UnderlyingAsset["TEL"] = "tel";
    UnderlyingAsset["TELEGRAMDAO"] = "telegramdao";
    UnderlyingAsset["TEMPO"] = "tempo";
    UnderlyingAsset["TEN"] = "ten";
    UnderlyingAsset["TENX"] = "tenx";
    UnderlyingAsset["TERC"] = "terc";
    UnderlyingAsset["TEUROC"] = "teuroc";
    UnderlyingAsset["TERC2DP"] = "terc2dp";
    UnderlyingAsset["TERC6DP"] = "terc6dp";
    UnderlyingAsset["TERC18DP"] = "terc18DP";
    UnderlyingAsset["TERC20"] = "terc20";
    UnderlyingAsset["TERC2DP1"] = "terc2dp1";
    UnderlyingAsset["TERC2DP2"] = "terc2dp2";
    UnderlyingAsset["TERC2DP3"] = "terc2dp3";
    UnderlyingAsset["TERC2DP4"] = "terc2dp4";
    UnderlyingAsset["TERC2DP5"] = "terc2dp5";
    UnderlyingAsset["TERC6DP1"] = "terc6dp1";
    UnderlyingAsset["TERC6DP2"] = "terc6dp2";
    UnderlyingAsset["TERC6DP3"] = "terc6dp3";
    UnderlyingAsset["TERC6DP4"] = "terc6dp4";
    UnderlyingAsset["TERC6DP5"] = "terc6dp5";
    UnderlyingAsset["TERC18DP1"] = "terc18dp1";
    UnderlyingAsset["TERC18DP2"] = "terc18dp2";
    UnderlyingAsset["TERC18DP3"] = "terc18dp3";
    UnderlyingAsset["TERC18DP4"] = "terc18dp4";
    UnderlyingAsset["TERC18DP5"] = "terc18dp5";
    UnderlyingAsset["TERC18DP6"] = "terc18dp6";
    UnderlyingAsset["TERC18DP7"] = "terc18dp7";
    UnderlyingAsset["TERC18DP8"] = "terc18dp8";
    UnderlyingAsset["TERC18DP9"] = "terc18dp9";
    UnderlyingAsset["TERC18DP10"] = "terc18dp10";
    UnderlyingAsset["TERC18DP11"] = "terc18dp11";
    UnderlyingAsset["TERC18DP12"] = "terc18dp12";
    UnderlyingAsset["TERC18DP13"] = "terc18dp13";
    UnderlyingAsset["TERC18DP14"] = "terc18dp14";
    UnderlyingAsset["TERC18DP15"] = "terc18dp15";
    UnderlyingAsset["BGERCH"] = "bgerch";
    UnderlyingAsset["AMSTEST"] = "amstest";
    UnderlyingAsset["TERM"] = "term";
    UnderlyingAsset["TGBP"] = "tgbp";
    UnderlyingAsset["TUSDS"] = "tusds";
    UnderlyingAsset["TGOUSD"] = "tgousd";
    UnderlyingAsset["hteth:gousd"] = "hteth:gousd";
    UnderlyingAsset["hteth:grtx"] = "hteth:grtx";
    UnderlyingAsset["hteth:sofid"] = "hteth:sofid";
    UnderlyingAsset["hteth:stgsofid"] = "hteth:stgsofid";
    UnderlyingAsset["hteth:usd1"] = "hteth:usd1";
    UnderlyingAsset["hteth:stgusd1"] = "hteth:stgusd1";
    UnderlyingAsset["hteth:cusd"] = "hteth:cusd";
    UnderlyingAsset["hteth:fyusd"] = "hteth:fyusd";
    UnderlyingAsset["hteth:stgcusd"] = "hteth:stgcusd";
    UnderlyingAsset["hteth:stgfyusd"] = "hteth:stgfyusd";
    UnderlyingAsset["hteth:tsteth"] = "hteth:tsteth";
    UnderlyingAsset["hteth:grtxp"] = "hteth:grtxp";
    UnderlyingAsset["hteth:testeigen"] = "hteth:testeigen";
    UnderlyingAsset["THKD"] = "thkd";
    UnderlyingAsset["THUNDER"] = "thunder";
    UnderlyingAsset["TIO"] = "tio";
    UnderlyingAsset["TIOX"] = "tiox";
    UnderlyingAsset["TKMK"] = "tkmk";
    UnderlyingAsset["TKNT"] = "tknt";
    UnderlyingAsset["TKO"] = "tko";
    UnderlyingAsset["TKX"] = "tkx";
    UnderlyingAsset["TLAB"] = "tlab";
    UnderlyingAsset["TLM"] = "tlm";
    UnderlyingAsset["TLOS"] = "tlos";
    UnderlyingAsset["TMATIC"] = "tmatic";
    UnderlyingAsset["TMSN"] = "tmsn";
    UnderlyingAsset["TNT"] = "tnt";
    UnderlyingAsset["TOKAMAK"] = "tokamak";
    UnderlyingAsset["TOKE"] = "toke";
    UnderlyingAsset["TOKEN"] = "token";
    UnderlyingAsset["TOMI"] = "tomi";
    UnderlyingAsset["TOMOBEAR"] = "tomobear";
    UnderlyingAsset["TOMOBEAR2"] = "tomobear2";
    UnderlyingAsset["TOMOBULL"] = "tomobull";
    UnderlyingAsset["TOK"] = "tok";
    UnderlyingAsset["TONCOIN"] = "toncoin";
    UnderlyingAsset["TOPM"] = "topm";
    UnderlyingAsset["TRAC"] = "trac";
    UnderlyingAsset["TRAXX"] = "traxx";
    UnderlyingAsset["TRB"] = "trb";
    UnderlyingAsset["TRIBE"] = "tribe";
    UnderlyingAsset["TRIBL"] = "tribl";
    UnderlyingAsset["TRL"] = "trl";
    UnderlyingAsset["TROY"] = "troy";
    UnderlyingAsset["TRST"] = "trst";
    UnderlyingAsset["TRU"] = "tru";
    UnderlyingAsset["TRUF"] = "truf";
    UnderlyingAsset["TRUFV2"] = "trufv2";
    UnderlyingAsset["TRUMPLOSE"] = "trumplose";
    UnderlyingAsset["TRUMPWIN"] = "trumpwin";
    UnderlyingAsset["TRXBEAR"] = "trxbear";
    UnderlyingAsset["TRXBULL"] = "trxbull";
    UnderlyingAsset["TRXDOOM"] = "trxdoom";
    UnderlyingAsset["TRX-ERC20"] = "TRX-ERC20";
    UnderlyingAsset["TRXHEDGE"] = "trxhedge";
    UnderlyingAsset["TRXMOON"] = "trxmoon";
    // Bilira
    UnderlyingAsset["TRYB"] = "tryb";
    // TRYB on Solana - https://solscan.io/token/6ry4WBDvAwAnrYJVv6MCog4J8zx6S3cPgSqnTsDZ73AR
    UnderlyingAsset["TRYB2"] = "tryb2";
    UnderlyingAsset["TRYBBEAR"] = "trybbear";
    UnderlyingAsset["TRYBBULL"] = "trybbull";
    UnderlyingAsset["TRYX"] = "tryx";
    UnderlyingAsset["TST"] = "tst";
    UnderlyingAsset["TSUKA"] = "tsuka";
    UnderlyingAsset["TULIP"] = "tulip";
    UnderlyingAsset["TUPOLIS"] = "tupolis";
    UnderlyingAsset["TUSD"] = "tusd";
    UnderlyingAsset["TUSDC"] = "tusdc";
    UnderlyingAsset["TUSDT"] = "tusdt";
    UnderlyingAsset["TUSRM"] = "tusrm";
    UnderlyingAsset["TWDOGE"] = "twdoge";
    UnderlyingAsset["TWETH"] = "tweth";
    UnderlyingAsset["TXL"] = "txl";
    UnderlyingAsset["TXSGD"] = "txsgd";
    UnderlyingAsset["TXUSD"] = "txusd";
    UnderlyingAsset["UAIR"] = "uair";
    UnderlyingAsset["UBXT"] = "ubxt";
    UnderlyingAsset["UCO"] = "uco";
    UnderlyingAsset["UFT"] = "uft";
    UnderlyingAsset["UKG"] = "ukg";
    UnderlyingAsset["UMA"] = "uma";
    UnderlyingAsset["UMEE"] = "umee";
    UnderlyingAsset["UNB"] = "unb";
    UnderlyingAsset["UNI"] = "uni";
    UnderlyingAsset["UOS"] = "uos";
    UnderlyingAsset["UP"] = "up";
    UnderlyingAsset["UPBTC"] = "upbtc";
    UnderlyingAsset["UPP"] = "upp";
    UnderlyingAsset["UPT"] = "upt";
    UnderlyingAsset["UPUSD"] = "upusd";
    UnderlyingAsset["UQC"] = "uqc";
    UnderlyingAsset["URHD"] = "urhd";
    UnderlyingAsset["sol:usdt"] = "sol:usdt";
    UnderlyingAsset["sol:usdc"] = "sol:usdc";
    UnderlyingAsset["sol:agri"] = "sol:agri";
    UnderlyingAsset["sol:usdca"] = "sol:usdca";
    UnderlyingAsset["USCC"] = "uscc";
    UnderlyingAsset["USDC"] = "usdc";
    UnderlyingAsset["USDC-POS-WORMHOLE"] = "usdc-pos-wormhole";
    UnderlyingAsset["USDD"] = "usdd";
    UnderlyingAsset["USDE"] = "usde";
    UnderlyingAsset["USDGLO"] = "usdglo";
    UnderlyingAsset["USDH"] = "usdh";
    UnderlyingAsset["USDK"] = "usdk";
    // Also available on EOS
    UnderlyingAsset["USDT"] = "usdt";
    UnderlyingAsset["USDTBEAR"] = "usdtbear";
    UnderlyingAsset["USDTBULL"] = "usdtbull";
    UnderlyingAsset["USDTDOOM"] = "usdtdoom";
    UnderlyingAsset["USDTHEDGE"] = "usdthedge";
    UnderlyingAsset["USDTMOON"] = "usdtmoon";
    UnderlyingAsset["USDX"] = "usdx";
    UnderlyingAsset["USDY"] = "usdy";
    UnderlyingAsset["USG"] = "usg";
    UnderlyingAsset["USPX"] = "uspx";
    UnderlyingAsset["UST"] = "ust";
    UnderlyingAsset["USTB"] = "ustb";
    UnderlyingAsset["UST-WORMHOLE"] = "ust-wormhole";
    UnderlyingAsset["USX"] = "usx";
    UnderlyingAsset["USYC"] = "usyc";
    UnderlyingAsset["UTK"] = "utk";
    UnderlyingAsset["UTK1"] = "utk1";
    UnderlyingAsset["UXB"] = "uxb";
    UnderlyingAsset["UXP"] = "uxp";
    UnderlyingAsset["VALOR"] = "valor";
    UnderlyingAsset["VANRY"] = "vanry";
    UnderlyingAsset["VBNT"] = "vbnt";
    UnderlyingAsset["VCORE"] = "vcore";
    UnderlyingAsset["VDX"] = "vdx";
    UnderlyingAsset["VEC"] = "vec";
    UnderlyingAsset["VEE"] = "vee";
    UnderlyingAsset["VEGA"] = "vega";
    UnderlyingAsset["VEXT"] = "vext";
    UnderlyingAsset["VGX"] = "vgx";
    UnderlyingAsset["VI"] = "vi";
    UnderlyingAsset["VIB"] = "vib";
    UnderlyingAsset["VIC"] = "vic";
    UnderlyingAsset["VIDT"] = "vidt";
    UnderlyingAsset["VISR"] = "visr";
    UnderlyingAsset["VIU"] = "viu";
    UnderlyingAsset["VOLT"] = "volt";
    UnderlyingAsset["VRA"] = "vra";
    UnderlyingAsset["VRGX"] = "vrgx";
    UnderlyingAsset["VRTX"] = "vrtx";
    UnderlyingAsset["VSP"] = "vsp";
    UnderlyingAsset["VXC"] = "vxc";
    UnderlyingAsset["VXV"] = "vxv";
    UnderlyingAsset["W"] = "w";
    // Wrapped AAVE
    UnderlyingAsset["WAAVE"] = "waave";
    UnderlyingAsset["WABI"] = "wabi";
    UnderlyingAsset["WAFL"] = "wafl";
    UnderlyingAsset["WAGMI"] = "wagmi";
    // Wrapped AAVAX
    UnderlyingAsset["WAVAX"] = "wavax";
    UnderlyingAsset["WAVES"] = "waves";
    UnderlyingAsset["WAX"] = "wax";
    UnderlyingAsset["WAXP"] = "waxp";
    // Wrapped BNB
    UnderlyingAsset["WBNB"] = "wbnb";
    UnderlyingAsset["WECAN"] = "wecan";
    UnderlyingAsset["WFEE"] = "wfee";
    UnderlyingAsset["WHAT"] = "what";
    UnderlyingAsset["WOO"] = "woo";
    UnderlyingAsset["WTK"] = "wtk";
    UnderlyingAsset["WBTC"] = "wbtc";
    UnderlyingAsset["WDAIV2"] = "wdaiv2";
    UnderlyingAsset["WDOGE"] = "wdoge";
    UnderlyingAsset["WCFG"] = "wcfg";
    UnderlyingAsset["WEC"] = "wec";
    UnderlyingAsset["sol:wec"] = "sol:wec";
    UnderlyingAsset["WET"] = "wet";
    UnderlyingAsset["WETH"] = "weth";
    UnderlyingAsset["WEETH"] = "weeth";
    UnderlyingAsset["WFLOW"] = "wflow";
    UnderlyingAsset["WFFT"] = "wfft";
    UnderlyingAsset["WHALE"] = "whale";
    UnderlyingAsset["WHT"] = "wht";
    UnderlyingAsset["WILD"] = "wild";
    UnderlyingAsset["WING"] = "wing";
    UnderlyingAsset["WNXM"] = "wnxm";
    UnderlyingAsset["WLD"] = "wld";
    UnderlyingAsset["WLUNA"] = "wluna";
    UnderlyingAsset["WLXT"] = "wlxt";
    // Wrapped SOL
    UnderlyingAsset["sol:wsol"] = "sol:wsol";
    // Wrapped Rose
    UnderlyingAsset["WROSE"] = "wrose";
    UnderlyingAsset["WSTETH"] = "wsteth";
    UnderlyingAsset["WPX"] = "wpx";
    UnderlyingAsset["WTAO"] = "wtao";
    UnderlyingAsset["WTC"] = "wtc";
    UnderlyingAsset["WTGXX"] = "wtgxx";
    // USD Coin (Wormhole)
    UnderlyingAsset["WUSDC"] = "wusdc";
    UnderlyingAsset["WUSDCV2"] = "wusdvcv2";
    UnderlyingAsset["WUSDM"] = "wusdm";
    // Tether USD (Wormhole)
    UnderlyingAsset["WUSDTV2"] = "wusdtv2";
    UnderlyingAsset["WXRP"] = "wxrp";
    UnderlyingAsset["WXRPV0"] = "wxrpv0";
    UnderlyingAsset["WXT"] = "wxt";
    UnderlyingAsset["XAUD"] = "xaud";
    UnderlyingAsset["XAURY"] = "xaury";
    UnderlyingAsset["XAUT"] = "xaut";
    UnderlyingAsset["XAUTBEAR"] = "xautbear";
    UnderlyingAsset["XAUTBULL"] = "xautbull";
    UnderlyingAsset["XBGOLD"] = "xbgold";
    UnderlyingAsset["XCD"] = "xcd";
    UnderlyingAsset["XCHNG"] = "xchng";
    UnderlyingAsset["XCN"] = "xcn";
    UnderlyingAsset["XDEFI"] = "xdefi";
    UnderlyingAsset["XDOGE"] = "xdoge";
    UnderlyingAsset["XEX"] = "xex";
    UnderlyingAsset["XLMBEAR"] = "xlmbear";
    UnderlyingAsset["XLMBULL"] = "xlmbull";
    UnderlyingAsset["xpl:syzusd"] = "xpl:syzusd";
    UnderlyingAsset["xpl:usdto"] = "xpl:usdto";
    UnderlyingAsset["XRL"] = "xrl";
    UnderlyingAsset["XRPBEAR"] = "xrpbear";
    UnderlyingAsset["XRPBULL"] = "xrpbull";
    UnderlyingAsset["XRPDOOM"] = "xrpdoom";
    UnderlyingAsset["XRPHEDGE"] = "xrphedge";
    UnderlyingAsset["XRPMOON"] = "xrpmoon";
    UnderlyingAsset["XSGD"] = "xsgd";
    UnderlyingAsset["XSUSHI"] = "xsushi";
    UnderlyingAsset["XTP"] = "xtp";
    UnderlyingAsset["XTZBEAR"] = "xtzbear";
    UnderlyingAsset["XTZBULL"] = "xtzbull";
    UnderlyingAsset["XUSD"] = "xusd";
    UnderlyingAsset["XVS"] = "xvs";
    UnderlyingAsset["XX"] = "xx";
    UnderlyingAsset["XZK"] = "xzk";
    UnderlyingAsset["YAMV2"] = "yamv2";
    UnderlyingAsset["YFDAI"] = "yfdai";
    UnderlyingAsset["YFI"] = "yfi";
    UnderlyingAsset["YFII"] = "yfii";
    UnderlyingAsset["YFL"] = "yfl";
    UnderlyingAsset["YGG"] = "ygg";
    UnderlyingAsset["YLD"] = "yld";
    UnderlyingAsset["YNG"] = "yng";
    UnderlyingAsset["YSEY"] = "ysey";
    UnderlyingAsset["ZARX"] = "zarx";
    UnderlyingAsset["ZBC"] = "zbc";
    UnderlyingAsset["ZBU"] = "zbu";
    UnderlyingAsset["ZBUV2"] = "zbuv2";
    UnderlyingAsset["ZCO"] = "zco";
    UnderlyingAsset["ZECBEAR"] = "zecbear";
    UnderlyingAsset["ZECBULL"] = "zecbull";
    UnderlyingAsset["ZETAEVM"] = "zetaevm";
    UnderlyingAsset["ZIL"] = "zil";
    UnderlyingAsset["ZIP"] = "zip";
    UnderlyingAsset["ZIX"] = "zix";
    UnderlyingAsset["ZKL"] = "zkl";
    UnderlyingAsset["ZKS"] = "zks";
    UnderlyingAsset["ZLW"] = "zlw";
    UnderlyingAsset["ZMT"] = "zmt";
    UnderlyingAsset["ZOOM"] = "zoom";
    UnderlyingAsset["ZRO"] = "zro";
    UnderlyingAsset["ZRO-0x320"] = "zro-0x320";
    UnderlyingAsset["ZRO-0xFCF"] = "zro-0xfcf";
    UnderlyingAsset["ZRO-0xE5C"] = "zro-0xe5c";
    UnderlyingAsset["ZRX"] = "zrx";
    UnderlyingAsset["ZUSD"] = "zusd";
    UnderlyingAsset["eth:usdg"] = "eth:usdg";
    UnderlyingAsset["eth:spxux"] = "eth:spxux";
    UnderlyingAsset["eth:aleo"] = "eth:aleo";
    UnderlyingAsset["eth:dbusd"] = "eth:dbusd";
    UnderlyingAsset["eth:edu"] = "eth:edu";
    UnderlyingAsset["eth:telos"] = "eth:telos";
    UnderlyingAsset["eth:cusdo"] = "eth:cusdo";
    UnderlyingAsset["eth:aevo"] = "eth:aevo";
    UnderlyingAsset["eth:alt"] = "eth:alt";
    UnderlyingAsset["eth:rtbl"] = "eth:rtbl";
    UnderlyingAsset["eth:virtual"] = "eth:virtual";
    UnderlyingAsset["eth:vice"] = "eth:vice";
    UnderlyingAsset["eth:audu"] = "eth:audu";
    UnderlyingAsset["eth:wlfi"] = "eth:wlfi";
    UnderlyingAsset["eth:kava"] = "eth:kava";
    UnderlyingAsset["eth:gousd"] = "eth:gousd";
    UnderlyingAsset["eth:iq"] = "eth:iq";
    UnderlyingAsset["eth:iris"] = "eth:iris";
    UnderlyingAsset["eth:hard"] = "eth:hard";
    UnderlyingAsset["eth:hegic"] = "eth:hegic";
    UnderlyingAsset["eth:spx"] = "eth:spx";
    UnderlyingAsset["eth:exrd"] = "eth:exrd";
    UnderlyingAsset["eth:turbo"] = "eth:turbo";
    UnderlyingAsset["eth:icnt"] = "eth:icnt";
    UnderlyingAsset["eth:god"] = "eth:god";
    UnderlyingAsset["eth:sky"] = "eth:sky";
    UnderlyingAsset["eth:uco"] = "eth:uco";
    UnderlyingAsset["eth:fuel"] = "eth:fuel";
    UnderlyingAsset["eth:xprism"] = "eth:xprism";
    UnderlyingAsset["eth:xreth"] = "eth:xreth";
    UnderlyingAsset["eth:xy"] = "eth:xy";
    UnderlyingAsset["eth:yu"] = "eth:yu";
    UnderlyingAsset["eth:move"] = "eth:move";
    UnderlyingAsset["eth:mon"] = "eth:mon";
    UnderlyingAsset["eth:usual"] = "eth:usual";
    UnderlyingAsset["eth:usd1"] = "eth:usd1";
    UnderlyingAsset["eth:usdm1"] = "eth:usdm1";
    UnderlyingAsset["eth:sofid"] = "eth:sofid";
    UnderlyingAsset["eth:cusd"] = "eth:cusd";
    UnderlyingAsset["eth:fyusd"] = "eth:fyusd";
    UnderlyingAsset["eth:ibtc"] = "eth:ibtc";
    UnderlyingAsset["eth:pyr"] = "eth:pyr";
    UnderlyingAsset["eth:una"] = "eth:una";
    UnderlyingAsset["eth:ads"] = "eth:ads";
    UnderlyingAsset["eth:fuelv1"] = "eth:fuelv1";
    UnderlyingAsset["eth:cet"] = "eth:cet";
    UnderlyingAsset["eth:unio"] = "eth:unio";
    UnderlyingAsset["eth:flttx"] = "eth:flttx";
    UnderlyingAsset["eth:wtsix"] = "eth:wtsix";
    UnderlyingAsset["eth:modrx"] = "eth:modrx";
    UnderlyingAsset["eth:techx"] = "eth:techx";
    UnderlyingAsset["eth:wtsyx"] = "eth:wtsyx";
    UnderlyingAsset["eth:wtlgx"] = "eth:wtlgx";
    UnderlyingAsset["eth:wttsx"] = "eth:wttsx";
    UnderlyingAsset["eth:tipsx"] = "eth:tipsx";
    UnderlyingAsset["eth:wtstx"] = "eth:wtstx";
    UnderlyingAsset["eth:lngvx"] = "eth:lngvx";
    UnderlyingAsset["eth:eqtyx"] = "eth:eqtyx";
    UnderlyingAsset["eth:deuro"] = "eth:deuro";
    UnderlyingAsset["eth:usat"] = "eth:usat";
    UnderlyingAsset["eth:usdf"] = "eth:usdf";
    UnderlyingAsset["eth:ausd"] = "eth:ausd";
    UnderlyingAsset["eth:ags"] = "eth:ags";
    UnderlyingAsset["eth:aus"] = "eth:aus";
    UnderlyingAsset["eth:reya"] = "eth:reya";
    UnderlyingAsset["eth:usdp"] = "eth:usdp";
    UnderlyingAsset["eth:grtx"] = "eth:grtx";
    UnderlyingAsset["eth:gaia"] = "eth:gaia";
    UnderlyingAsset["eth:usds"] = "eth:usds";
    UnderlyingAsset["eth:perc"] = "eth:perc";
    UnderlyingAsset["eth:cfg"] = "eth:cfg";
    UnderlyingAsset["eth:plume"] = "eth:plume";
    UnderlyingAsset["eth:vbill"] = "eth:vbill";
    UnderlyingAsset["eth:la"] = "eth:la";
    UnderlyingAsset["eth:es"] = "eth:es";
    UnderlyingAsset["eth:ctrl"] = "eth:ctrl";
    UnderlyingAsset["eth:benji"] = "eth:benji";
    UnderlyingAsset["eth:ibenji"] = "eth:ibenji";
    UnderlyingAsset["eth:chex"] = "eth:chex";
    UnderlyingAsset["eth:gho"] = "eth:gho";
    UnderlyingAsset["eth:npc"] = "eth:npc";
    UnderlyingAsset["eth:towns"] = "eth:towns";
    UnderlyingAsset["eth:umint"] = "eth:umint";
    UnderlyingAsset["eth:arb"] = "eth:arb";
    UnderlyingAsset["eth:ez"] = "eth:ez";
    UnderlyingAsset["eth:ncash"] = "eth:ncash";
    UnderlyingAsset["eth:sub"] = "eth:sub";
    UnderlyingAsset["eth:poe"] = "eth:poe";
    UnderlyingAsset["eth:ocn"] = "eth:ocn";
    UnderlyingAsset["eth:banca"] = "eth:banca";
    UnderlyingAsset["eth:stq"] = "eth:stq";
    UnderlyingAsset["eth:route"] = "eth:route";
    UnderlyingAsset["eth:ryt"] = "eth:ryt";
    UnderlyingAsset["eth:guild"] = "eth:guild";
    UnderlyingAsset["eth:rdo"] = "eth:rdo";
    UnderlyingAsset["eth:h"] = "eth:h";
    UnderlyingAsset["eth:wbt"] = "eth:wbt";
    UnderlyingAsset["eth:ftn"] = "eth:ftn";
    UnderlyingAsset["eth:sc"] = "eth:sc";
    UnderlyingAsset["eth:lf"] = "eth:lf";
    UnderlyingAsset["eth:usdcv"] = "eth:usdcv";
    UnderlyingAsset["eth:cake"] = "eth:cake";
    UnderlyingAsset["eth:nft"] = "eth:nft";
    UnderlyingAsset["eth:morpho"] = "eth:morpho";
    UnderlyingAsset["eth:usdd"] = "eth:usdd";
    UnderlyingAsset["eth:mx"] = "eth:mx";
    UnderlyingAsset["eth:flz"] = "eth:flz";
    UnderlyingAsset["eth:usd0"] = "eth:usd0";
    UnderlyingAsset["eth:white"] = "eth:white";
    UnderlyingAsset["eth:upc"] = "eth:upc";
    UnderlyingAsset["eth:lgct"] = "eth:lgct";
    UnderlyingAsset["eth:usdtb"] = "eth:usdtb";
    UnderlyingAsset["eth:deusd"] = "eth:deusd";
    UnderlyingAsset["eth:neiro"] = "eth:neiro";
    UnderlyingAsset["eth:vana"] = "eth:vana";
    UnderlyingAsset["eth:eurau"] = "eth:eurau";
    UnderlyingAsset["eth:insur"] = "eth:insur";
    UnderlyingAsset["eth:xyo"] = "eth:xyo";
    UnderlyingAsset["eth:zig"] = "eth:zig";
    UnderlyingAsset["eth:swftc"] = "eth:swftc";
    UnderlyingAsset["eth:dsync"] = "eth:dsync";
    UnderlyingAsset["eth:orbr"] = "eth:orbr";
    UnderlyingAsset["eth:sxt"] = "eth:sxt";
    UnderlyingAsset["eth:paal"] = "eth:paal";
    UnderlyingAsset["eth:wmtx"] = "eth:wmtx";
    UnderlyingAsset["eth:anime"] = "eth:anime";
    UnderlyingAsset["eth:newt"] = "eth:newt";
    UnderlyingAsset["eth:hsk"] = "eth:hsk";
    UnderlyingAsset["eth:rog"] = "eth:rog";
    UnderlyingAsset["eth:xaum"] = "eth:xaum";
    UnderlyingAsset["eth:avail"] = "eth:avail";
    UnderlyingAsset["eth:dolo"] = "eth:dolo";
    UnderlyingAsset["eth:era"] = "eth:era";
    UnderlyingAsset["eth:ugold"] = "eth:ugold";
    UnderlyingAsset["eth:seda"] = "eth:seda";
    UnderlyingAsset["eth:enso"] = "eth:enso";
    UnderlyingAsset["eth:hpp"] = "eth:hpp";
    UnderlyingAsset["eth:lit"] = "eth:lit";
    UnderlyingAsset["eth:aedz"] = "eth:aedz";
    UnderlyingAsset["eth:arm-susde-usde"] = "eth:arm-susde-usde";
    UnderlyingAsset["eth:arm-weth-eeth"] = "eth:arm-weth-eeth";
    UnderlyingAsset["eth:cashplus"] = "eth:cashplus";
    UnderlyingAsset["eth:island"] = "eth:island";
    UnderlyingAsset["eth:six"] = "eth:six";
    UnderlyingAsset["eth:eden"] = "eth:eden";
    UnderlyingAsset["eth:xeden"] = "eth:xeden";
    UnderlyingAsset["eth:linea"] = "eth:linea";
    UnderlyingAsset["eth:ff"] = "eth:ff";
    UnderlyingAsset["eth:mavia"] = "eth:mavia";
    UnderlyingAsset["eth:lm"] = "eth:lm";
    UnderlyingAsset["eth:kub"] = "eth:kub";
    UnderlyingAsset["eth:fidd"] = "eth:fidd";
    UnderlyingAsset["eth:meme"] = "eth:meme";
    UnderlyingAsset["eth:bard"] = "eth:bard";
    UnderlyingAsset["eth:sfp"] = "eth:sfp";
    UnderlyingAsset["eth:aztec"] = "eth:aztec";
    // Ondo Tokenized Assets
    UnderlyingAsset["eth:qqqon"] = "qqqon";
    UnderlyingAsset["eth:spyon"] = "spyon";
    UnderlyingAsset["eth:nvdaon"] = "nvdaon";
    UnderlyingAsset["eth:tslaon"] = "tslaon";
    UnderlyingAsset["eth:aaplon"] = "aaplon";
    UnderlyingAsset["eth:mstron"] = "mstron";
    UnderlyingAsset["eth:pltron"] = "pltron";
    UnderlyingAsset["eth:hoodon"] = "hoodon";
    UnderlyingAsset["eth:crclon"] = "crclon";
    UnderlyingAsset["eth:coinon"] = "coinon";
    UnderlyingAsset["eth:amznon"] = "amznon";
    UnderlyingAsset["eth:googlon"] = "googlon";
    UnderlyingAsset["eth:metaon"] = "metaon";
    UnderlyingAsset["eth:babaon"] = "babaon";
    UnderlyingAsset["eth:msfton"] = "msfton";
    UnderlyingAsset["eth:spgion"] = "spgion";
    UnderlyingAsset["eth:tsmon"] = "tsmon";
    UnderlyingAsset["eth:amdon"] = "amdon";
    UnderlyingAsset["eth:unhon"] = "unhon";
    UnderlyingAsset["eth:jpmon"] = "jpmon";
    UnderlyingAsset["eth:orclon"] = "orclon";
    UnderlyingAsset["eth:von"] = "von";
    UnderlyingAsset["eth:maon"] = "maon";
    UnderlyingAsset["eth:llyon"] = "llyon";
    UnderlyingAsset["eth:nflxon"] = "nflxon";
    UnderlyingAsset["eth:coston"] = "coston";
    UnderlyingAsset["eth:iauon"] = "iauon";
    UnderlyingAsset["eth:ivvon"] = "ivvon";
    UnderlyingAsset["eth:slvon"] = "slvon";
    UnderlyingAsset["eth:iwnon"] = "eth:iwnon";
    UnderlyingAsset["eth:qbtson"] = "eth:qbtson";
    UnderlyingAsset["eth:tipon"] = "eth:tipon";
    UnderlyingAsset["eth:ulon"] = "eth:ulon";
    UnderlyingAsset["eth:itoton"] = "eth:itoton";
    UnderlyingAsset["eth:hygon"] = "eth:hygon";
    UnderlyingAsset["eth:gmeon"] = "eth:gmeon";
    UnderlyingAsset["eth:pbron"] = "eth:pbron";
    UnderlyingAsset["eth:eqixon"] = "eth:eqixon";
    UnderlyingAsset["eth:rioton"] = "eth:rioton";
    UnderlyingAsset["eth:maraon"] = "eth:maraon";
    UnderlyingAsset["eth:pfeon"] = "eth:pfeon";
    UnderlyingAsset["eth:eemon"] = "eth:eemon";
    UnderlyingAsset["eth:cmcsaon"] = "eth:cmcsaon";
    UnderlyingAsset["eth:sonyon"] = "eth:sonyon";
    UnderlyingAsset["eth:pyplon"] = "eth:pyplon";
    UnderlyingAsset["eth:himson"] = "eth:himson";
    UnderlyingAsset["eth:abton"] = "eth:abton";
    UnderlyingAsset["eth:nvoon"] = "eth:nvoon";
    UnderlyingAsset["eth:efaon"] = "eth:efaon";
    UnderlyingAsset["eth:nkeon"] = "eth:nkeon";
    UnderlyingAsset["eth:intcon"] = "eth:intcon";
    UnderlyingAsset["eth:iefaon"] = "eth:iefaon";
    UnderlyingAsset["eth:smcion"] = "eth:smcion";
    UnderlyingAsset["eth:mrvlon"] = "eth:mrvlon";
    UnderlyingAsset["eth:aggon"] = "eth:aggon";
    UnderlyingAsset["eth:sbuxon"] = "eth:sbuxon";
    UnderlyingAsset["eth:uberon"] = "eth:uberon";
    UnderlyingAsset["eth:abnbon"] = "eth:abnbon";
    UnderlyingAsset["eth:qcomon"] = "eth:qcomon";
    UnderlyingAsset["eth:cscoon"] = "eth:cscoon";
    UnderlyingAsset["eth:futuon"] = "eth:futuon";
    UnderlyingAsset["eth:cmgon"] = "eth:cmgon";
    UnderlyingAsset["eth:iwfon"] = "eth:iwfon";
    UnderlyingAsset["eth:koon"] = "eth:koon";
    UnderlyingAsset["eth:linon"] = "eth:linon";
    UnderlyingAsset["eth:armon"] = "eth:armon";
    UnderlyingAsset["eth:jdon"] = "eth:jdon";
    UnderlyingAsset["eth:muon"] = "eth:muon";
    UnderlyingAsset["eth:wmton"] = "eth:wmton";
    UnderlyingAsset["eth:tmon"] = "eth:tmon";
    UnderlyingAsset["eth:shopon"] = "eth:shopon";
    UnderlyingAsset["eth:rddton"] = "eth:rddton";
    UnderlyingAsset["eth:dison"] = "eth:dison";
    UnderlyingAsset["eth:apoon"] = "eth:apoon";
    UnderlyingAsset["eth:pepon"] = "eth:pepon";
    UnderlyingAsset["eth:wfcon"] = "eth:wfcon";
    UnderlyingAsset["eth:biduon"] = "eth:biduon";
    UnderlyingAsset["eth:mson"] = "eth:mson";
    UnderlyingAsset["eth:pgon"] = "eth:pgon";
    UnderlyingAsset["eth:cvxon"] = "eth:cvxon";
    UnderlyingAsset["eth:panwon"] = "eth:panwon";
    UnderlyingAsset["eth:avgoon"] = "eth:avgoon";
    UnderlyingAsset["eth:crmon"] = "eth:crmon";
    UnderlyingAsset["eth:snowon"] = "eth:snowon";
    UnderlyingAsset["eth:axpon"] = "eth:axpon";
    UnderlyingAsset["eth:ibmon"] = "eth:ibmon";
    UnderlyingAsset["eth:dashon"] = "eth:dashon";
    UnderlyingAsset["eth:acnon"] = "eth:acnon";
    UnderlyingAsset["eth:ijhon"] = "eth:ijhon";
    UnderlyingAsset["eth:baon"] = "eth:baon";
    UnderlyingAsset["eth:geon"] = "eth:geon";
    UnderlyingAsset["eth:appon"] = "eth:appon";
    UnderlyingAsset["eth:lmton"] = "eth:lmton";
    UnderlyingAsset["eth:intuon"] = "eth:intuon";
    UnderlyingAsset["eth:mcdon"] = "eth:mcdon";
    UnderlyingAsset["eth:gson"] = "eth:gson";
    UnderlyingAsset["eth:adbeon"] = "eth:adbeon";
    UnderlyingAsset["eth:spoton"] = "eth:spoton";
    UnderlyingAsset["eth:blkon"] = "eth:blkon";
    UnderlyingAsset["eth:asmlon"] = "eth:asmlon";
    UnderlyingAsset["eth:nowon"] = "eth:nowon";
    UnderlyingAsset["eth:iwmon"] = "eth:iwmon";
    UnderlyingAsset["eth:melion"] = "eth:melion";
    UnderlyingAsset["eth:tlton"] = "eth:tlton";
    UnderlyingAsset["eth:grndon"] = "eth:grndon";
    UnderlyingAsset["eth:figon"] = "eth:figon";
    UnderlyingAsset["eth:iemgon"] = "eth:iemgon";
    UnderlyingAsset["eth:sbeton"] = "eth:sbeton";
    UnderlyingAsset["eth:usdo"] = "eth:usdo";
    UnderlyingAsset["eth:align"] = "eth:align";
    UnderlyingAsset["eth:xan"] = "eth:xan";
    UnderlyingAsset["eth:frxusd"] = "eth:frxusd";
    UnderlyingAsset["eth:red"] = "eth:red";
    UnderlyingAsset["eth:dka"] = "eth:dka";
    UnderlyingAsset["eth:cgpt"] = "eth:cgpt";
    UnderlyingAsset["eth:apu"] = "eth:apu";
    UnderlyingAsset["eth:shfl"] = "eth:shfl";
    UnderlyingAsset["eth:banana"] = "eth:banana";
    UnderlyingAsset["eth:zkj"] = "eth:zkj";
    UnderlyingAsset["eth:spk"] = "eth:spk";
    UnderlyingAsset["eth:merl"] = "eth:merl";
    UnderlyingAsset["eth:aeur"] = "eth:aeur";
    UnderlyingAsset["eth:soso"] = "eth:soso";
    UnderlyingAsset["eth:bfc"] = "eth:bfc";
    UnderlyingAsset["eth:osak"] = "eth:osak";
    UnderlyingAsset["eth:uds"] = "eth:uds";
    UnderlyingAsset["eth:zent"] = "eth:zent";
    UnderlyingAsset["eth:euri"] = "eth:euri";
    UnderlyingAsset["eth:al"] = "eth:al";
    UnderlyingAsset["eth:wct"] = "eth:wct";
    UnderlyingAsset["eth:pundiai"] = "eth:pundiai";
    UnderlyingAsset["eth:anon"] = "eth:anon";
    UnderlyingAsset["eth:omi"] = "eth:omi";
    UnderlyingAsset["eth:andy"] = "eth:andy";
    UnderlyingAsset["eth:aioz"] = "eth:aioz";
    UnderlyingAsset["eth:job"] = "eth:job";
    UnderlyingAsset["eth:irys"] = "eth:irys";
    UnderlyingAsset["eth:kpk"] = "eth:kpk";
    UnderlyingAsset["eth:devve"] = "eth:devve";
    UnderlyingAsset["eth:fbtc"] = "eth:fbtc";
    UnderlyingAsset["eth:byzusd"] = "eth:byzusd";
    UnderlyingAsset["eth:audm"] = "eth:audm";
    UnderlyingAsset["eth:usdi"] = "eth:usdi";
    UnderlyingAsset["eth:tea"] = "eth:tea";
    UnderlyingAsset["eth:ofc"] = "eth:ofc";
    UnderlyingAsset["eth:wxm"] = "eth:wxm";
    UnderlyingAsset["eth:jpyc"] = "eth:jpyc";
    UnderlyingAsset["eth:ten"] = "eth:ten";
    UnderlyingAsset["eth:camp"] = "eth:camp";
    UnderlyingAsset["eth:f"] = "eth:f";
    UnderlyingAsset["eth:turtle"] = "eth:turtle";
    UnderlyingAsset["eth:order"] = "eth:order";
    UnderlyingAsset["eth:puffer"] = "eth:puffer";
    UnderlyingAsset["eth:resolv"] = "eth:resolv";
    UnderlyingAsset["eth:spec"] = "eth:spec";
    UnderlyingAsset["eth:prompt"] = "eth:prompt";
    UnderlyingAsset["eth:yb"] = "eth:yb";
    UnderlyingAsset["morph:usdc"] = "morph:usdc";
    UnderlyingAsset["morpheth:usdc"] = "morpheth:usdc";
    UnderlyingAsset["morph:usdt"] = "morph:usdt";
    UnderlyingAsset["morpheth:usdt"] = "morpheth:usdt";
    UnderlyingAsset["morph:usd1"] = "morph:usd1";
    UnderlyingAsset["morpheth:usd1"] = "morpheth:usd1";
    UnderlyingAsset["tmorph:tmt"] = "tmorph:tmt";
    UnderlyingAsset["tmorpheth:tmt"] = "tmorpheth:tmt";
    UnderlyingAsset["tmorpheth:usd1"] = "tmorpheth:usd1";
    UnderlyingAsset["tmorpheth:stgusd1"] = "tmorpheth:stgusd1";
    UnderlyingAsset["xlm:BST-GADDFE4R72YUP2AOEL67OHZN3GJQYPC3VE734N2XFMEGRR2L32CZ3XYZ"] = "xlm:BST-GADDFE4R72YUP2AOEL67OHZN3GJQYPC3VE734N2XFMEGRR2L32CZ3XYZ";
    UnderlyingAsset["xlm:VELO-GDM4RQUQQUVSKQA7S6EM7XBZP3FCGH4Q7CL6TABQ7B2BEJ5ERARM2M5M"] = "xlm:VELO-GDM4RQUQQUVSKQA7S6EM7XBZP3FCGH4Q7CL6TABQ7B2BEJ5ERARM2M5M";
    UnderlyingAsset["xlm:SLT-GCKA6K5PCQ6PNF5RQBF7PQDJWRHO6UOGFMRLK3DYHDOI244V47XKQ4GP"] = "xlm:SLT-GCKA6K5PCQ6PNF5RQBF7PQDJWRHO6UOGFMRLK3DYHDOI244V47XKQ4GP";
    UnderlyingAsset["xlm:USD-GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX"] = "xlm:USD-GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX";
    UnderlyingAsset["xlm:ETH-GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5"] = "xlm:ETH-GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5";
    UnderlyingAsset["xlm:WXT-GASBLVHS5FOABSDNW5SPPH3QRJYXY5JHA2AOA2QHH2FJLZBRXSG4SWXT"] = "xlm:WXT-GASBLVHS5FOABSDNW5SPPH3QRJYXY5JHA2AOA2QHH2FJLZBRXSG4SWXT";
    UnderlyingAsset["xlm:USDC-GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN"] = "xlm:USDC-GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN";
    UnderlyingAsset["xlm:USDM1-GDM5QWWXCMDTQMZAKMYTCI52LA7FWBHAZMU5NJLMIFHDJISJRP2ZWPKC"] = "xlm:USDM1-GDM5QWWXCMDTQMZAKMYTCI52LA7FWBHAZMU5NJLMIFHDJISJRP2ZWPKC";
    UnderlyingAsset["xlm:SIX-GDMS6EECOH6MBMCP3FYRYEVRBIV3TQGLOFQIPVAITBRJUMTI6V7A2X6Z"] = "xlm:SIX-GDMS6EECOH6MBMCP3FYRYEVRBIV3TQGLOFQIPVAITBRJUMTI6V7A2X6Z";
    UnderlyingAsset["xlm:BRLT-GCHQ3F2BF5P74DMDNOOGHT5DUCKC773AW5DTOFINC26W4KGYFPYDPRSO"] = "xlm:BRLT-GCHQ3F2BF5P74DMDNOOGHT5DUCKC773AW5DTOFINC26W4KGYFPYDPRSO";
    UnderlyingAsset["xlm:ARST-GCSAZVWXZKWS4XS223M5F54H2B6XPIIXZZGP7KEAIU6YSL5HDRGCI3DG"] = "xlm:ARST-GCSAZVWXZKWS4XS223M5F54H2B6XPIIXZZGP7KEAIU6YSL5HDRGCI3DG";
    UnderlyingAsset["xlm:AQUA-GBNZILSTVQZ4R7IKQDGHYGY2QXL5QOFJYQMXPKWRRM5PAV7Y4M67AQUA"] = "xlm:AQUA-GBNZILSTVQZ4R7IKQDGHYGY2QXL5QOFJYQMXPKWRRM5PAV7Y4M67AQUA";
    UnderlyingAsset["xlm:EURC-GDHU6WRG4IEQXM5NZ4BMPKOXHW76MZM4Y2IEMFDVXBSDP6SJY4ITNPP2"] = "xlm:EURC-GDHU6WRG4IEQXM5NZ4BMPKOXHW76MZM4Y2IEMFDVXBSDP6SJY4ITNPP2";
    UnderlyingAsset["xlm:GYEN-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB"] = "xlm:GYEN-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB";
    UnderlyingAsset["xlm:ZUSD-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB"] = "xlm:ZUSD-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB";
    UnderlyingAsset["xlm:EURS-GC5FGCDEOGOGSNWCCNKS3OMEVDHTE3Q5A5FEQWQKV3AXA7N6KDQ2CUZJ"] = "xlm:EURS-GC5FGCDEOGOGSNWCCNKS3OMEVDHTE3Q5A5FEQWQKV3AXA7N6KDQ2CUZJ";
    UnderlyingAsset["xlm:VEUR-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN"] = "xlm:VEUR-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN";
    UnderlyingAsset["xlm:VCHF-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN"] = "xlm:VCHF-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN";
    UnderlyingAsset["xlm:AUDD-GDC7X2MXTYSAKUUGAIQ7J7RPEIM7GXSAIWFYWWH4GLNFECQVJJLB2EEU"] = "xlm:AUDD-GDC7X2MXTYSAKUUGAIQ7J7RPEIM7GXSAIWFYWWH4GLNFECQVJJLB2EEU";
    UnderlyingAsset["xlm:BENJI-GBHNGLLIE3KWGKCHIKMHJ5HVZHYIK7WTBE4QF5PLAKL4CJGSEU7HZIW5"] = "xlm:BENJI-GBHNGLLIE3KWGKCHIKMHJ5HVZHYIK7WTBE4QF5PLAKL4CJGSEU7HZIW5";
    UnderlyingAsset["xlm:gBENJI-GD5J73EKK5IYL5XS3FBTHHX7CZIYRP7QXDL57XFWGC2WVYWT326OBXRP"] = "xlm:gBENJI-GD5J73EKK5IYL5XS3FBTHHX7CZIYRP7QXDL57XFWGC2WVYWT326OBXRP";
    UnderlyingAsset["xlm:SHX-GDSTRSHXHGJ7ZIVRBXEYE5Q74XUVCUSEKEBR7UCHEUUEK72N7I7KJ6JH"] = "xlm:SHX-GDSTRSHXHGJ7ZIVRBXEYE5Q74XUVCUSEKEBR7UCHEUUEK72N7I7KJ6JH";
    // Eth NFTs
    // generic NFTs
    UnderlyingAsset["erc721:token"] = "erc721:token";
    UnderlyingAsset["erc1155:token"] = "erc1155:token";
    UnderlyingAsset["nonstandard:token"] = "nonstandard:token";
    // Test Eth NFTs
    UnderlyingAsset["terc721:token"] = "terc721:token";
    UnderlyingAsset["terc1155:token"] = "terc1155:token";
    UnderlyingAsset["tnonstandard:token"] = "tnonstandard:token";
    // Algorand mainnet tokens
    UnderlyingAsset["algo:USDC-31566704"] = "algo:USDC-31566704";
    UnderlyingAsset["algo:USDt-312769"] = "algo:USDt-312769";
    UnderlyingAsset["algo:MCAU-6547014"] = "algo:MCAU-6547014";
    UnderlyingAsset["algo:QCAD-84507107"] = "algo:QCAD-84507107";
    UnderlyingAsset["algo:VCAD-438505559"] = "algo:VCAD-438505559";
    // Kovan-only ERC20 tokens
    UnderlyingAsset["TEST"] = "test";
    UnderlyingAsset["SCHZ"] = "schz";
    UnderlyingAsset["CAT"] = "cat";
    // Stellar testnet tokens
    UnderlyingAsset["txlm:BST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L"] = "txlm:BST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L";
    UnderlyingAsset["txlm:TST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L"] = "txlm:TST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L";
    // Algorand testnet tokens
    UnderlyingAsset["talgo:USON-16026728"] = "talgo:USON-16026728";
    UnderlyingAsset["talgo:SPRW-16026732"] = "talgo:SPRW-16026732";
    UnderlyingAsset["talgo:KAL-16026733"] = "talgo:KAL-16026733";
    UnderlyingAsset["talgo:USDC-10458941"] = "talgo:USDC-10458941";
    UnderlyingAsset["talgo:USDt-180447"] = "talgo:USDt-180447";
    UnderlyingAsset["talgo:JPT-162085446"] = "talgo:JPT-162085446";
    // EOS tokens
    UnderlyingAsset["CHEX"] = "chex";
    UnderlyingAsset["IQ"] = "iq";
    UnderlyingAsset["EOS_BOX"] = "eos:box";
    UnderlyingAsset["EOS_SBET"] = "eos:sbet";
    UnderlyingAsset["VAULTA"] = "vaulta";
    // Avax Token ERC-20
    UnderlyingAsset["avaxc:qi"] = "avaxc:qi";
    UnderlyingAsset["avaxc:xava"] = "avaxc:xava";
    UnderlyingAsset["avaxc:klo"] = "avaxc:klo";
    UnderlyingAsset["avaxc:joe"] = "avaxc:joe";
    UnderlyingAsset["avaxc:png"] = "avaxc:png";
    UnderlyingAsset["avaxc:usdt"] = "avaxc:usdt";
    UnderlyingAsset["avaxc:usdc"] = "avaxc:usdc";
    UnderlyingAsset["avaxc:link"] = "avaxc:link";
    UnderlyingAsset["avaxc:cai"] = "avaxc:cai";
    UnderlyingAsset["avaxc:aave"] = "avaxc:aave";
    UnderlyingAsset["avaxc:btc"] = "avaxc:btc";
    UnderlyingAsset["avaxc:dai"] = "avaxc:dai";
    UnderlyingAsset["avaxc:arena"] = "avaxc:arena";
    UnderlyingAsset["avaxc:tryb"] = "avaxc:tryb";
    UnderlyingAsset["avaxc:wbtc"] = "avaxc:wbtc";
    UnderlyingAsset["avaxc:weth"] = "avaxc:weth";
    UnderlyingAsset["avaxc:sbc"] = "avaxc:sbc";
    UnderlyingAsset["avaxc:xsgd"] = "avaxc:xsgd";
    UnderlyingAsset["avaxc:ticov2"] = "avaxc:ticov2";
    UnderlyingAsset["avaxc:nxpc"] = "avaxc:nxpc";
    UnderlyingAsset["avaxc:spxux"] = "avaxc:spxux";
    UnderlyingAsset["avaxc:stavax"] = "avaxc:stavax";
    UnderlyingAsset["tavaxc:opm"] = "tavaxc:opm";
    UnderlyingAsset["tavaxc:cop2peq"] = "tavaxc:cop2peq";
    UnderlyingAsset["tavaxc:xsgd"] = "tavaxc:xsgd";
    UnderlyingAsset["tavaxc:bitgo"] = "tavaxc:bitgo";
    UnderlyingAsset["tavaxc:stavax"] = "tavaxc:stavax";
    UnderlyingAsset["avaxc:usdc-e"] = "avaxc:usdc-e";
    UnderlyingAsset["avaxc:usdt-e"] = "avaxc:usdt-e";
    // Begin FTX missing AVAXC tokens
    UnderlyingAsset["avaxc:yeti"] = "avaxc:yeti";
    UnderlyingAsset["avaxc:spell"] = "avaxc:spell";
    UnderlyingAsset["avaxc:yusd"] = "avaxc:yusd";
    UnderlyingAsset["avaxc:yusdcrv-f"] = "avaxc:yusdcrv-f";
    UnderlyingAsset["avaxc:ecd"] = "avaxc:ecd";
    UnderlyingAsset["avaxc:blzz"] = "avaxc:blzz";
    UnderlyingAsset["avaxc:ptp"] = "avaxc:ptp";
    UnderlyingAsset["avaxc:stg"] = "avaxc:stg";
    UnderlyingAsset["avaxc:syn"] = "avaxc:syn";
    UnderlyingAsset["avaxc:aavausdc"] = "avaxc:aavausdc";
    UnderlyingAsset["avaxc:tusd"] = "avaxc:tusd";
    UnderlyingAsset["avaxc:crv"] = "avaxc:crv";
    UnderlyingAsset["avaxc:savax"] = "avaxc:savax";
    UnderlyingAsset["avaxc:ampl"] = "avaxc:ampl";
    UnderlyingAsset["avaxc:cnr"] = "avaxc:cnr";
    UnderlyingAsset["avaxc:roco"] = "avaxc:roco";
    UnderlyingAsset["avaxc:aavadai"] = "avaxc:aavadai";
    UnderlyingAsset["avaxc:vtx"] = "avaxc:vtx";
    UnderlyingAsset["avaxc:wavax"] = "avaxc:wavax";
    UnderlyingAsset["avaxc:bnb"] = "avaxc:bnb";
    UnderlyingAsset["avaxc:aavausdt"] = "avaxc:aavausdt";
    UnderlyingAsset["avaxc:acre"] = "avaxc:acre";
    UnderlyingAsset["avaxc:gmx"] = "avaxc:gmx";
    UnderlyingAsset["avaxc:gunz"] = "avaxc:gunz";
    UnderlyingAsset["avaxc:mim"] = "avaxc:mim";
    UnderlyingAsset["avaxc:axlusdc"] = "avaxc:axlusdc";
    UnderlyingAsset["avaxc:lot"] = "avaxc:lot";
    UnderlyingAsset["avaxc:av3crv"] = "avaxc:av3crv";
    UnderlyingAsset["avaxc:time"] = "avaxc:time";
    UnderlyingAsset["avaxc:uni.e"] = "avaxc:uni.e";
    UnderlyingAsset["avaxc:sb"] = "avaxc:sb";
    UnderlyingAsset["avaxc:dyp"] = "avaxc:dyp";
    UnderlyingAsset["avaxc:sing"] = "avaxc:sing";
    UnderlyingAsset["avaxc:gohm"] = "avaxc:gohm";
    UnderlyingAsset["avaxc:boofi"] = "avaxc:boofi";
    UnderlyingAsset["avaxc:eth"] = "avaxc:eth";
    UnderlyingAsset["avaxc:wmemo"] = "avaxc:wmemo";
    UnderlyingAsset["avaxc:fxs"] = "avaxc:fxs";
    UnderlyingAsset["avaxc:sifu"] = "avaxc:sifu";
    UnderlyingAsset["avaxc:sushi.e"] = "avaxc:sushi.e";
    UnderlyingAsset["avaxc:sushi"] = "avaxc:sushi";
    UnderlyingAsset["avaxc:mimatic"] = "avaxc:mimatic";
    UnderlyingAsset["avaxc:sspell"] = "avaxc:sspell";
    UnderlyingAsset["avaxc:grape"] = "avaxc:grape";
    UnderlyingAsset["avaxc:xjoe"] = "avaxc:xjoe";
    UnderlyingAsset["avaxc:bsgg"] = "avaxc:bsgg";
    UnderlyingAsset["avaxc:roy"] = "avaxc:roy";
    UnderlyingAsset["avaxc:wow"] = "avaxc:wow";
    UnderlyingAsset["avaxc:wine"] = "avaxc:wine";
    UnderlyingAsset["avaxc:mu"] = "avaxc:mu";
    UnderlyingAsset["avaxc:frax"] = "avaxc:frax";
    UnderlyingAsset["avaxc:movr"] = "avaxc:movr";
    UnderlyingAsset["avaxc:ice"] = "avaxc:ice";
    UnderlyingAsset["avaxc:note"] = "avaxc:note";
    UnderlyingAsset["avaxc:wrose"] = "avaxc:wrose";
    UnderlyingAsset["avaxc:swap"] = "avaxc:swap";
    UnderlyingAsset["avaxc:tico"] = "avaxc:tico";
    UnderlyingAsset["avaxc:shrap"] = "avaxc:shrap";
    UnderlyingAsset["avaxc:benji"] = "avaxc:benji";
    UnderlyingAsset["avaxc:emdx"] = "avaxc:emdx";
    UnderlyingAsset["avaxc:eurc"] = "avaxc:eurc";
    // End FTX missing AVAXC tokens
    // polygon Token ERC-20
    UnderlyingAsset["polygon:usdc"] = "polygon:usdc";
    UnderlyingAsset["polygon:usdcv2"] = "polygon:usdcv2";
    UnderlyingAsset["polygon:usdt"] = "polygon:usdt";
    UnderlyingAsset["polygon:weth"] = "polygon:weth";
    UnderlyingAsset["polygon:cnkt"] = "polygon:cnkt";
    UnderlyingAsset["polygon:wbtc"] = "polygon:wbtc";
    UnderlyingAsset["polygon:sand"] = "polygon:sand";
    UnderlyingAsset["polygon:dai"] = "polygon:dai";
    UnderlyingAsset["polygon:woo"] = "polygon:woo";
    UnderlyingAsset["polygon:aave"] = "polygon:aave";
    UnderlyingAsset["polygon:link"] = "polygon:link";
    UnderlyingAsset["polygon:tusd"] = "polygon:tusd";
    UnderlyingAsset["polygon:cel"] = "polygon:cel";
    UnderlyingAsset["polygon:busd"] = "polygon:busd";
    UnderlyingAsset["polygon:frax"] = "polygon:frax";
    UnderlyingAsset["polygon:crv"] = "polygon:crv";
    UnderlyingAsset["polygon:uni"] = "polygon:uni";
    UnderlyingAsset["polygon:fcd"] = "polygon:fcd";
    UnderlyingAsset["polygon:ape"] = "polygon:ape";
    UnderlyingAsset["polygon:srm"] = "polygon:srm";
    UnderlyingAsset["polygon:fly"] = "polygon:fly";
    UnderlyingAsset["polygon:gfc"] = "polygon:gfc";
    UnderlyingAsset["polygon:rbw"] = "polygon:rbw";
    UnderlyingAsset["polygon:zed"] = "polygon:zed";
    UnderlyingAsset["polygon:vext"] = "polygon:vext";
    UnderlyingAsset["polygon:vcnt"] = "polygon:vcnt";
    UnderlyingAsset["polygon:sushi"] = "polygon:sushi";
    UnderlyingAsset["polygon:wmatic"] = "polygon:wmatic";
    UnderlyingAsset["polygon:1inch"] = "polygon:1inch";
    UnderlyingAsset["polygon:comp"] = "polygon:comp";
    UnderlyingAsset["polygon:sol"] = "polygon:sol";
    UnderlyingAsset["polygon:wavax"] = "polygon:wavax";
    UnderlyingAsset["polygon:wbnb"] = "polygon:wbnb";
    UnderlyingAsset["polygon:wftm"] = "polygon:wftm";
    UnderlyingAsset["polygon:yfi"] = "polygon:yfi";
    UnderlyingAsset["polygon:treta"] = "polygon:treta";
    UnderlyingAsset["polygon:orb"] = "polygon:orb";
    UnderlyingAsset["polygon:route"] = "polygon:route";
    UnderlyingAsset["polygon:sbc"] = "polygon:sbc";
    UnderlyingAsset["polygon:xsgd"] = "polygon:xsgd";
    UnderlyingAsset["polygon:dimo"] = "polygon:dimo";
    UnderlyingAsset["polygon:bcut"] = "polygon:bcut";
    UnderlyingAsset["polygon:pme"] = "polygon:pme";
    UnderlyingAsset["polygon:dipe"] = "polygon:dipe";
    UnderlyingAsset["polygon:lif3"] = "polygon:lif3";
    UnderlyingAsset["polygon:l3usd"] = "polygon:l3usd";
    UnderlyingAsset["polygon:moca"] = "polygon:moca";
    UnderlyingAsset["polygon:mask"] = "polygon:mask";
    UnderlyingAsset["polygon:nexo"] = "polygon:nexo";
    UnderlyingAsset["polygon:om"] = "polygon:om";
    UnderlyingAsset["polygon:pyr"] = "polygon:pyr";
    UnderlyingAsset["polygon:renbtc"] = "polygon:renbtc";
    UnderlyingAsset["polygon:req"] = "polygon:req";
    UnderlyingAsset["polygon:rndr"] = "polygon:rndr";
    UnderlyingAsset["polygon:snx"] = "polygon:snx";
    UnderlyingAsset["polygon:trb"] = "polygon:trb";
    UnderlyingAsset["polygon:ali"] = "polygon:ali";
    UnderlyingAsset["polygon:bal"] = "polygon:bal";
    UnderlyingAsset["polygon:elon"] = "polygon:elon";
    UnderlyingAsset["polygon:hex"] = "polygon:hex";
    UnderlyingAsset["polygon:iotx"] = "polygon:iotx";
    UnderlyingAsset["polygon:agix"] = "polygon:agix";
    UnderlyingAsset["polygon:avax"] = "polygon:avax";
    UnderlyingAsset["polygon:band"] = "polygon:band";
    UnderlyingAsset["polygon:blz"] = "polygon:blz";
    UnderlyingAsset["polygon:bnb"] = "polygon:bnb";
    UnderlyingAsset["polygon:bnt"] = "polygon:bnt";
    UnderlyingAsset["polygon:chz"] = "polygon:chz";
    UnderlyingAsset["polygon:enj"] = "polygon:enj";
    UnderlyingAsset["polygon:fet"] = "polygon:fet";
    UnderlyingAsset["polygon:forth"] = "polygon:forth";
    UnderlyingAsset["polygon:glm"] = "polygon:glm";
    UnderlyingAsset["polygon:gno"] = "polygon:gno";
    UnderlyingAsset["polygon:gohm"] = "polygon:gohm";
    UnderlyingAsset["polygon:gtc"] = "polygon:gtc";
    UnderlyingAsset["polygon:gusd"] = "polygon:gusd";
    UnderlyingAsset["polygon:hot"] = "polygon:hot";
    UnderlyingAsset["polygon:inj"] = "polygon:inj";
    UnderlyingAsset["polygon:lit"] = "polygon:lit";
    UnderlyingAsset["polygon:lrc"] = "polygon:lrc";
    UnderlyingAsset["polygon:mana"] = "polygon:mana";
    UnderlyingAsset["polygon:shib"] = "polygon:shib";
    UnderlyingAsset["polygon:sxp"] = "polygon:sxp";
    UnderlyingAsset["polygon:grt"] = "polygon:grt";
    UnderlyingAsset["polygon:mkr"] = "polygon:mkr";
    UnderlyingAsset["polygon:oxt"] = "polygon:oxt";
    UnderlyingAsset["polygon:pax"] = "polygon:pax";
    UnderlyingAsset["polygon:paxg"] = "polygon:paxg";
    UnderlyingAsset["polygon:powr"] = "polygon:powr";
    UnderlyingAsset["polygon:super"] = "polygon:super";
    UnderlyingAsset["polygon:uma"] = "polygon:uma";
    UnderlyingAsset["polygon:zrx"] = "polygon:zrx";
    UnderlyingAsset["polygon:ont"] = "polygon:ont";
    UnderlyingAsset["polygon:wrx"] = "polygon:wrx";
    UnderlyingAsset["polygon:voxel"] = "polygon:voxel";
    UnderlyingAsset["polygon:uft"] = "polygon:uft";
    UnderlyingAsset["polygon:ooki"] = "polygon:ooki";
    UnderlyingAsset["polygon:swap"] = "polygon:swap";
    UnderlyingAsset["polygon:vanry"] = "polygon:vanry";
    UnderlyingAsset["polygon:npt"] = "polygon:npt";
    UnderlyingAsset["polygon:volt"] = "polygon:volt";
    UnderlyingAsset["polygon:euroe"] = "polygon:euroe";
    UnderlyingAsset["polygon:geod"] = "polygon:geod";
    UnderlyingAsset["polygon:heth"] = "polygon:heth";
    UnderlyingAsset["polygon:copm"] = "polygon:copm";
    UnderlyingAsset["polygon:gmt"] = "polygon:gmt";
    UnderlyingAsset["polygon:uhu"] = "polygon:uhu";
    UnderlyingAsset["polygon:mv"] = "polygon:mv";
    UnderlyingAsset["polygon:bid"] = "polygon:bid";
    UnderlyingAsset["polygon:tcs"] = "polygon:tcs";
    UnderlyingAsset["polygon:buidl"] = "polygon:buidl";
    UnderlyingAsset["polygon:benji"] = "polygon:benji";
    UnderlyingAsset["polygon:naka"] = "polygon:naka";
    UnderlyingAsset["polygon:xusd"] = "polygon:xusd";
    UnderlyingAsset["polygon:txusd"] = "polygon:txusd";
    UnderlyingAsset["polygon:zig"] = "polygon:zig";
    UnderlyingAsset["polygon:brl1"] = "polygon:brl1";
    UnderlyingAsset["polygon:cnktplus"] = "polygon:cnktplus";
    UnderlyingAsset["polygon:land1"] = "polygon:land1";
    UnderlyingAsset["polygon:mmt54"] = "polygon:mmt54";
    UnderlyingAsset["polygon:jpyc"] = "polygon:jpyc";
    UnderlyingAsset["polygon:argt"] = "polygon:argt";
    UnderlyingAsset["polygon:brat"] = "polygon:brat";
    UnderlyingAsset["polygon:chlt"] = "polygon:chlt";
    UnderlyingAsset["polygon:stg"] = "polygon:stg";
    // Polygon NFTs
    // generic NFTs
    UnderlyingAsset["erc721:polygontoken"] = "erc721:polygontoken";
    UnderlyingAsset["erc1155:polygontoken"] = "erc1155:polygontoken";
    // BSC Token BEP-20
    UnderlyingAsset["bsc:sol"] = "bsc:sol";
    UnderlyingAsset["bsc:solv"] = "bsc:solv";
    UnderlyingAsset["bsc:stable"] = "bsc:stable";
    UnderlyingAsset["bsc:brise"] = "bsc:brise";
    UnderlyingAsset["bsc:bsw"] = "bsc:bsw";
    UnderlyingAsset["bsc:burger"] = "bsc:burger";
    UnderlyingAsset["bsc:cfx"] = "bsc:cfx";
    UnderlyingAsset["bsc:bake"] = "bsc:bake";
    UnderlyingAsset["bsc:bnx"] = "bsc:bnx";
    UnderlyingAsset["bsc:busd"] = "bsc:busd";
    UnderlyingAsset["bsc:hook"] = "bsc:hook";
    UnderlyingAsset["bsc:ksm"] = "bsc:ksm";
    UnderlyingAsset["bsc:usdt"] = "bsc:usdt";
    UnderlyingAsset["bsc:vet"] = "bsc:vet";
    UnderlyingAsset["bsc:cake"] = "bsc:cake";
    UnderlyingAsset["bsc:litt"] = "bsc:litt";
    UnderlyingAsset["bsc:xvs"] = "bsc:xvs";
    UnderlyingAsset["bsc:epx"] = "bsc:epx";
    UnderlyingAsset["bsc:usdc"] = "bsc:usdc";
    UnderlyingAsset["bsc:eth"] = "bsc:eth";
    UnderlyingAsset["bsc:dd"] = "bsc:dd";
    UnderlyingAsset["bsc:parti"] = "bsc:parti";
    UnderlyingAsset["bsc:form"] = "bsc:form";
    UnderlyingAsset["bsc:ake"] = "bsc:ake";
    UnderlyingAsset["bsc:kgen"] = "bsc:kgen";
    UnderlyingAsset["bsc:myx"] = "bsc:myx";
    UnderlyingAsset["bsc:ltc"] = "bsc:ltc";
    UnderlyingAsset["bsc:mask"] = "bsc:mask";
    UnderlyingAsset["bsc:matic"] = "bsc:matic";
    UnderlyingAsset["bsc:mbox"] = "bsc:mbox";
    UnderlyingAsset["bsc:mdt"] = "bsc:mdt";
    UnderlyingAsset["bsc:nuls"] = "bsc:nuls";
    UnderlyingAsset["bsc:ont"] = "bsc:ont";
    UnderlyingAsset["bsc:orn"] = "bsc:orn";
    UnderlyingAsset["bsc:porto"] = "bsc:porto";
    UnderlyingAsset["bsc:reef"] = "bsc:reef";
    UnderlyingAsset["bsc:renbtc"] = "bsc:renbtc";
    UnderlyingAsset["bsc:snx"] = "bsc:snx";
    UnderlyingAsset["bsc:tking"] = "bsc:tking";
    UnderlyingAsset["bsc:tlm"] = "bsc:tlm";
    UnderlyingAsset["bsc:ton"] = "bsc:ton";
    UnderlyingAsset["bsc:trx"] = "bsc:trx";
    UnderlyingAsset["bsc:wbnb"] = "bsc:wbnb";
    UnderlyingAsset["bsc:win"] = "bsc:win";
    UnderlyingAsset["bsc:wrx"] = "bsc:wrx";
    UnderlyingAsset["bsc:yfii"] = "bsc:yfii";
    UnderlyingAsset["bsc:zil"] = "bsc:zil";
    UnderlyingAsset["bsc:1inch"] = "bsc:1inch";
    UnderlyingAsset["bsc:ada"] = "bsc:ada";
    UnderlyingAsset["bsc:alice"] = "bsc:alice";
    UnderlyingAsset["bsc:alpaca"] = "bsc:alpaca";
    UnderlyingAsset["bsc:alpine"] = "bsc:alpine";
    UnderlyingAsset["bsc:ankr"] = "bsc:ankr";
    UnderlyingAsset["bsc:avax"] = "bsc:avax";
    UnderlyingAsset["bsc:beta"] = "bsc:beta";
    UnderlyingAsset["bsc:btt"] = "bsc:btt";
    UnderlyingAsset["bsc:celr"] = "bsc:celr";
    UnderlyingAsset["bsc:chr"] = "bsc:chr";
    UnderlyingAsset["bsc:coti"] = "bsc:coti";
    UnderlyingAsset["bsc:cream"] = "bsc:cream";
    UnderlyingAsset["bsc:dar"] = "bsc:dar";
    UnderlyingAsset["bsc:degov2"] = "bsc:degov2";
    UnderlyingAsset["bsc:dodo"] = "bsc:dodo";
    UnderlyingAsset["bsc:elon"] = "bsc:elon";
    UnderlyingAsset["bsc:etc"] = "bsc:etc";
    UnderlyingAsset["bsc:firo"] = "bsc:firo";
    UnderlyingAsset["bsc:front"] = "bsc:front";
    UnderlyingAsset["bsc:hft"] = "bsc:hft";
    UnderlyingAsset["bsc:high"] = "bsc:high";
    UnderlyingAsset["bsc:inj"] = "bsc:inj";
    UnderlyingAsset["bsc:iotx"] = "bsc:iotx";
    UnderlyingAsset["bsc:auto"] = "bsc:auto";
    UnderlyingAsset["bsc:fet"] = "bsc:fet";
    UnderlyingAsset["bsc:kas"] = "bsc:kas";
    UnderlyingAsset["bsc:lit"] = "bsc:lit";
    UnderlyingAsset["bsc:mana"] = "bsc:mana";
    UnderlyingAsset["bsc:shib"] = "bsc:shib";
    UnderlyingAsset["bsc:sxp"] = "bsc:sxp";
    UnderlyingAsset["bsc:nnn"] = "bsc:nnn";
    UnderlyingAsset["bsc:nvm"] = "bsc:nvm";
    UnderlyingAsset["bsc:jasmy"] = "bsc:jasmy";
    UnderlyingAsset["bsc:near"] = "bsc:near";
    UnderlyingAsset["bsc:ocean"] = "bsc:ocean";
    UnderlyingAsset["bsc:sand"] = "bsc:sand";
    UnderlyingAsset["bsc:tusd"] = "bsc:tusd";
    UnderlyingAsset["bsc:wrose"] = "bsc:wrose";
    UnderlyingAsset["bsc:twt"] = "bsc:twt";
    UnderlyingAsset["bsc:sfp"] = "bsc:sfp";
    UnderlyingAsset["bsc:edu"] = "bsc:edu";
    UnderlyingAsset["bsc:mrs"] = "bsc:mrs";
    UnderlyingAsset["bsc:ong"] = "bsc:ong";
    UnderlyingAsset["bsc:ctk"] = "bsc:ctk";
    UnderlyingAsset["bsc:rndt"] = "bsc:rndt";
    UnderlyingAsset["bsc:mbx"] = "bsc:mbx";
    UnderlyingAsset["bsc:mav"] = "bsc:mav";
    UnderlyingAsset["bsc:mct"] = "bsc:mct";
    UnderlyingAsset["bsc:thunder"] = "bsc:thunder";
    UnderlyingAsset["bsc:atlas"] = "bsc:atlas";
    UnderlyingAsset["bsc:vidt"] = "bsc:vidt";
    UnderlyingAsset["bsc:unfi"] = "bsc:unfi";
    UnderlyingAsset["bsc:chess"] = "bsc:chess";
    UnderlyingAsset["bsc:pols"] = "bsc:pols";
    UnderlyingAsset["bsc:uft"] = "bsc:uft";
    UnderlyingAsset["bsc:wing"] = "bsc:wing";
    UnderlyingAsset["bsc:santos"] = "bsc:santos";
    UnderlyingAsset["bsc:lazio"] = "bsc:lazio";
    UnderlyingAsset["bsc:swap"] = "bsc:swap";
    UnderlyingAsset["bsc:troy"] = "bsc:troy";
    UnderlyingAsset["bsc:rdnt"] = "bsc:rdnt";
    UnderlyingAsset["bsc:pax"] = "bsc:pax";
    UnderlyingAsset["bsc:volt"] = "bsc:volt";
    UnderlyingAsset["tbsc:busd"] = "tbsc:busd";
    UnderlyingAsset["tbsc:usd1"] = "tbsc:usd1";
    UnderlyingAsset["tbsc:stgusd1"] = "tbsc:stgusd1";
    UnderlyingAsset["bsc:city"] = "bsc:city";
    UnderlyingAsset["bsc:fdusd"] = "bsc:fdusd";
    UnderlyingAsset["bsc:floki"] = "bsc:floki";
    UnderlyingAsset["bsc:ldo"] = "bsc:ldo";
    UnderlyingAsset["bsc:om"] = "bsc:om";
    UnderlyingAsset["bsc:eos"] = "bsc:eos";
    UnderlyingAsset["bsc:usdd"] = "bsc:usdd";
    UnderlyingAsset["bsc:gft"] = "bsc:gft";
    UnderlyingAsset["bsc:glmr"] = "bsc:glmr";
    UnderlyingAsset["bsc:gmt"] = "bsc:gmt";
    UnderlyingAsset["bsc:tko"] = "bsc:tko";
    UnderlyingAsset["bsc:vite"] = "bsc:vite";
    UnderlyingAsset["bsc:mdx"] = "bsc:mdx";
    UnderlyingAsset["bsc:multi"] = "bsc:multi";
    UnderlyingAsset["bsc:psg"] = "bsc:psg";
    UnderlyingAsset["bsc:telos"] = "bsc:telos";
    UnderlyingAsset["bsc:flux"] = "bsc:flux";
    UnderlyingAsset["bsc:h2o"] = "bsc:h2o";
    UnderlyingAsset["bsc:lto"] = "bsc:lto";
    UnderlyingAsset["bsc:kmd"] = "bsc:kmd";
    UnderlyingAsset["bsc:farm"] = "bsc:farm";
    UnderlyingAsset["bsc:lina"] = "bsc:lina";
    UnderlyingAsset["bsc:usd1"] = "bsc:usd1";
    UnderlyingAsset["bsc:oort"] = "bsc:oort";
    UnderlyingAsset["bsc:aitech"] = "bsc:aitech";
    UnderlyingAsset["bsc:fil"] = "bsc:fil";
    UnderlyingAsset["bsc:ftm"] = "bsc:ftm";
    UnderlyingAsset["bsc:comp"] = "bsc:comp";
    UnderlyingAsset["bsc:uni"] = "bsc:uni";
    UnderlyingAsset["bsc:yfi"] = "bsc:yfi";
    UnderlyingAsset["bsc:link"] = "bsc:link";
    UnderlyingAsset["bsc:cusdo"] = "bsc:cusdo";
    UnderlyingAsset["bsc:unx"] = "bsc:unx";
    UnderlyingAsset["bsc:usdo"] = "bsc:usdo";
    UnderlyingAsset["bsc:slay"] = "bsc:slay";
    UnderlyingAsset["bsc:prove"] = "bsc:prove";
    UnderlyingAsset["bsc:rekt"] = "bsc:rekt";
    UnderlyingAsset["bsc:zig"] = "bsc:zig";
    UnderlyingAsset["bsc:eden"] = "bsc:eden";
    UnderlyingAsset["bsc:m"] = "bsc:m";
    UnderlyingAsset["bsc:cashplus"] = "bsc:cashplus";
    UnderlyingAsset["bsc:aster"] = "bsc:aster";
    UnderlyingAsset["bsc:pln"] = "bsc:pln";
    UnderlyingAsset["bsc:ff"] = "bsc:ff";
    UnderlyingAsset["bsc:c98"] = "bsc:c98";
    UnderlyingAsset["bsc:ln"] = "bsc:ln";
    UnderlyingAsset["bsc:soon"] = "bsc:soon";
    UnderlyingAsset["bsc:bard"] = "bsc:bard";
    UnderlyingAsset["bsc:home"] = "bsc:home";
    UnderlyingAsset["bsc:zbt"] = "bsc:zbt";
    UnderlyingAsset["bsc:iost"] = "bsc:iost";
    UnderlyingAsset["bsc:sto"] = "bsc:sto";
    UnderlyingAsset["bsc:pt-cusdo-29oct2026"] = "bsc:pt-cusdo-29oct2026";
    // BSC NFTs
    // generic NFTs
    UnderlyingAsset["erc721:bsctoken"] = "erc721:bsctoken";
    UnderlyingAsset["erc1155:bsctoken"] = "erc1155:bsctoken";
    // Test BSC NFTs
    UnderlyingAsset["terc721:bsctoken"] = "terc721:bsctoken";
    UnderlyingAsset["terc1155:bsctoken"] = "terc1155:bsctoken";
    // Polygon testnet tokens
    UnderlyingAsset["tpolygon:derc20"] = "tpolygon:derc20";
    UnderlyingAsset["tpolygon:link"] = "tpolygon:link";
    UnderlyingAsset["tpolygon:name"] = "tpolygon:name";
    UnderlyingAsset["tpolygon:opm"] = "tpolygon:opm";
    UnderlyingAsset["tpolygon:pme"] = "tpolygon:pme";
    UnderlyingAsset["tpolygon:xsgd"] = "tpolygon:xsgd";
    UnderlyingAsset["tpolygon:terc18dp"] = "tpolygon:terc18dp";
    UnderlyingAsset["tpolygon:terc10dp"] = "tpolygon:terc10dp";
    UnderlyingAsset["tpolygon:terc6dp"] = "tpolygon:terc6dp";
    UnderlyingAsset["tpolygon:usdt"] = "tpolygon:usdt";
    UnderlyingAsset["tpolygon:usdc"] = "tpolygon:usdc";
    UnderlyingAsset["tpolygon:testcopm"] = "tpolygon:testcopm";
    UnderlyingAsset["tpolygon:BitGoTest"] = "tpolygon:BitGoTest";
    UnderlyingAsset["tpolygon:copr3"] = "tpolygon:copr3";
    // generic NFTs
    UnderlyingAsset["terc721:polygontoken"] = "terc721:polygontoken";
    UnderlyingAsset["terc1155:polygontoken"] = "terc1155:polygontoken";
    // Arbitrum mainnet tokens
    UnderlyingAsset["arbeth:link"] = "arbeth:link";
    UnderlyingAsset["arbeth:spxux"] = "arbeth:spxux";
    UnderlyingAsset["arbeth:usdc"] = "arbeth:usdc";
    UnderlyingAsset["arbeth:xsgdv2"] = "arbeth:xsgdv2";
    UnderlyingAsset["arbeth:trn"] = "arbeth:trn";
    UnderlyingAsset["arbeth:usdcv2"] = "arbeth:usdcv2";
    UnderlyingAsset["arbeth:usdt"] = "arbeth:usdt";
    UnderlyingAsset["arbeth:arb"] = "arbeth:arb";
    UnderlyingAsset["arbeth:sqd"] = "arbeth:sqd";
    UnderlyingAsset["arbeth:cbl"] = "arbeth:cbl";
    UnderlyingAsset["arbeth:w"] = "arbeth:w";
    UnderlyingAsset["arbeth:comp"] = "arbeth:comp";
    UnderlyingAsset["arbeth:coti"] = "arbeth:coti";
    UnderlyingAsset["arbeth:gno"] = "arbeth:gno";
    UnderlyingAsset["arbeth:gohm"] = "arbeth:gohm";
    UnderlyingAsset["arbeth:grt"] = "arbeth:grt";
    UnderlyingAsset["arbeth:knc"] = "arbeth:knc";
    UnderlyingAsset["arbeth:trb"] = "arbeth:trb";
    UnderlyingAsset["arbeth:tusd"] = "arbeth:tusd";
    UnderlyingAsset["arbeth:uma"] = "arbeth:uma";
    UnderlyingAsset["arbeth:uni"] = "arbeth:uni";
    UnderlyingAsset["arbeth:weth"] = "arbeth:weth";
    UnderlyingAsset["arbeth:woo"] = "arbeth:woo";
    UnderlyingAsset["arbeth:yfi"] = "arbeth:yfi";
    UnderlyingAsset["arbeth:xsgd"] = "arbeth:xsgd";
    UnderlyingAsset["arbeth:ztx"] = "arbeth:ztx";
    UnderlyingAsset["arbeth:ldo"] = "arbeth:ldo";
    UnderlyingAsset["arbeth:egp"] = "arbeth:egp";
    UnderlyingAsset["arbeth:myrc"] = "arbeth:myrc";
    UnderlyingAsset["arbeth:gs"] = "arbeth:gs";
    UnderlyingAsset["arbeth:veur"] = "arbeth:veur";
    UnderlyingAsset["arbeth:vchf"] = "arbeth:vchf";
    UnderlyingAsset["arbeth:tbill"] = "arbeth:tbill";
    UnderlyingAsset["arbeth:xai"] = "arbeth:xai";
    UnderlyingAsset["arbeth:flttx"] = "arbeth:flttx";
    UnderlyingAsset["arbeth:wtsix"] = "arbeth:wtsix";
    UnderlyingAsset["arbeth:modrx"] = "arbeth:modrx";
    UnderlyingAsset["arbeth:techx"] = "arbeth:techx";
    UnderlyingAsset["arbeth:wtsyx"] = "arbeth:wtsyx";
    UnderlyingAsset["arbeth:wtlgx"] = "arbeth:wtlgx";
    UnderlyingAsset["arbeth:wttsx"] = "arbeth:wttsx";
    UnderlyingAsset["arbeth:tipsx"] = "arbeth:tipsx";
    UnderlyingAsset["arbeth:wtstx"] = "arbeth:wtstx";
    UnderlyingAsset["arbeth:wtgxx"] = "arbeth:wtgxx";
    UnderlyingAsset["arbeth:lngvx"] = "arbeth:lngvx";
    UnderlyingAsset["arbeth:eqtyx"] = "arbeth:eqtyx";
    UnderlyingAsset["arbeth:anime"] = "arbeth:anime";
    UnderlyingAsset["arbeth:benji"] = "arbeth:benji";
    UnderlyingAsset["arbeth:dolo"] = "arbeth:dolo";
    UnderlyingAsset["arbeth:bull"] = "arbeth:bull";
    UnderlyingAsset["arbeth:rdnt"] = "arbeth:rdnt";
    UnderlyingAsset["arbeth:pendle"] = "arbeth:pendle";
    UnderlyingAsset["arbeth:gmx"] = "arbeth:gmx";
    UnderlyingAsset["arbeth:uxlink"] = "arbeth:uxlink";
    UnderlyingAsset["arbeth:next"] = "arbeth:next";
    UnderlyingAsset["arbeth:zro"] = "arbeth:zro";
    UnderlyingAsset["arbeth:tt"] = "arbeth:tt";
    // BaseETH mainnet tokens
    UnderlyingAsset["baseeth:aero"] = "baseeth:aero";
    UnderlyingAsset["baseeth:cfi"] = "baseeth:cfi";
    UnderlyingAsset["baseeth:icnt"] = "baseeth:icnt";
    UnderlyingAsset["baseeth:mey"] = "baseeth:mey";
    UnderlyingAsset["baseeth:morpho"] = "baseeth:morpho";
    UnderlyingAsset["baseeth:myrc"] = "baseeth:myrc";
    UnderlyingAsset["baseeth:weth"] = "baseeth:weth";
    UnderlyingAsset["baseeth:usdc"] = "baseeth:usdc";
    UnderlyingAsset["baseeth:wbtc"] = "baseeth:wbtc";
    UnderlyingAsset["baseeth:usde"] = "baseeth:usde";
    UnderlyingAsset["baseeth:trust"] = "baseeth:trust";
    UnderlyingAsset["baseeth:flk"] = "baseeth:flk";
    UnderlyingAsset["baseeth:soon"] = "baseeth:soon";
    UnderlyingAsset["baseeth:wave"] = "baseeth:wave";
    UnderlyingAsset["baseeth:spec"] = "baseeth:spec";
    UnderlyingAsset["baseeth:tig"] = "baseeth:tig";
    UnderlyingAsset["baseeth:virtual"] = "baseeth:virtual";
    UnderlyingAsset["baseeth:zora"] = "baseeth:zora";
    UnderlyingAsset["baseeth:toshi"] = "baseeth:toshi";
    UnderlyingAsset["baseeth:creator"] = "baseeth:creator";
    UnderlyingAsset["baseeth:avnt"] = "baseeth:avnt";
    UnderlyingAsset["baseeth:mira"] = "baseeth:mira";
    UnderlyingAsset["baseeth:towns"] = "baseeth:towns";
    UnderlyingAsset["baseeth:brlv"] = "baseeth:brlv";
    UnderlyingAsset["baseeth:wbrly"] = "baseeth:wbrly";
    UnderlyingAsset["baseeth:recall"] = "baseeth:recall";
    UnderlyingAsset["baseeth:sapien"] = "baseeth:sapien";
    UnderlyingAsset["baseeth:aixbt"] = "baseeth:aixbt";
    UnderlyingAsset["baseeth:brett"] = "baseeth:brett";
    UnderlyingAsset["baseeth:argt"] = "baseeth:argt";
    UnderlyingAsset["baseeth:brat"] = "baseeth:brat";
    UnderlyingAsset["baseeth:mext"] = "baseeth:mext";
    UnderlyingAsset["baseeth:common"] = "baseeth:common";
    UnderlyingAsset["baseeth:unite"] = "baseeth:unite";
    UnderlyingAsset["baseeth:b3"] = "baseeth:b3";
    UnderlyingAsset["baseeth:kaito"] = "baseeth:kaito";
    // BaseETH testnet tokens
    UnderlyingAsset["tbaseeth:usdc"] = "tbaseeth:usdc";
    UnderlyingAsset["tbaseeth:xusd"] = "tbaseeth:xusd";
    UnderlyingAsset["tbaseeth:xsgd"] = "tbaseeth:xsgd";
    UnderlyingAsset["tbaseeth:usd1cx"] = "tbaseeth:usd1cx";
    UnderlyingAsset["tbaseeth:ctusd1cx"] = "tbaseeth:ctusd1cx";
    // Og mainnet tokens
    UnderlyingAsset["og:wog"] = "og:wog";
    // Seievm mainnet tokens
    UnderlyingAsset["seievm:usdc"] = "seievm:usdc";
    UnderlyingAsset["seievm:weth"] = "seievm:weth";
    UnderlyingAsset["seievm:wbtc"] = "seievm:wbtc";
    UnderlyingAsset["seievm:usd0"] = "seievm:usd0";
    UnderlyingAsset["seievm:wsei"] = "seievm:wsei";
    //Linea mainnet tokens
    UnderlyingAsset["lineaeth:linea"] = "lineaeth:linea";
    UnderlyingAsset["lineaeth:usdt"] = "lineaeth:usdt";
    UnderlyingAsset["lineaeth:usdc"] = "lineaeth:usdc";
    // Mantle mainnet tokens
    UnderlyingAsset["mantle:usdt"] = "mantle:usdt";
    UnderlyingAsset["mantle:usdc"] = "mantle:usdc";
    UnderlyingAsset["mantle:usde"] = "mantle:usde";
    UnderlyingAsset["mantle:usdt0"] = "mantle:usdt0";
    UnderlyingAsset["mantle:ausd"] = "mantle:ausd";
    UnderlyingAsset["mantle:usd1"] = "mantle:usd1";
    // Mantle testnet tokens
    UnderlyingAsset["tmantle:bgerch"] = "tmantle:bgerch";
    UnderlyingAsset["tmantle:usd1"] = "tmantle:usd1";
    UnderlyingAsset["tmantle:stgusd1"] = "tmantle:stgusd1";
    // Flow mainnet tokens
    UnderlyingAsset["flow:weth"] = "flow:weth";
    UnderlyingAsset["flow:usdf"] = "flow:usdf";
    UnderlyingAsset["flow:wflow"] = "flow:wflow";
    // Monad mainnet tokens
    UnderlyingAsset["mon:usdc"] = "mon:usdc";
    UnderlyingAsset["mon:wmon"] = "mon:wmon";
    // Monad testnet tokens
    UnderlyingAsset["tmon:tmt"] = "tmon:tmt";
    // XDC mainnet tokens
    UnderlyingAsset["xdc:usdc"] = "xdc:usdc";
    UnderlyingAsset["xdc:lbt"] = "xdc:lbt";
    UnderlyingAsset["xdc:cre"] = "xdc:cre";
    UnderlyingAsset["xdc:gama"] = "xdc:gama";
    UnderlyingAsset["xdc:srx"] = "xdc:srx";
    UnderlyingAsset["xdc:weth"] = "xdc:weth";
    // XDC testnet tokens
    UnderlyingAsset["txdc:tmt"] = "txdc:tmt";
    // hypeeevm testnet tokens
    UnderlyingAsset["thypeevm:usdc"] = "thypeevm:usdc";
    // hypeevm mainnet tokens
    UnderlyingAsset["hypeevm:hwhype"] = "hypeevm:hwhype";
    UnderlyingAsset["hypeevm:usdc"] = "hypeevm:usdc";
    UnderlyingAsset["hypeevm:usdt0"] = "hypeevm:usdt0";
    // Jovayeth mainnet tokens
    UnderlyingAsset["jovayeth:jft"] = "jovayeth:jft";
    UnderlyingAsset["jovayeth:usdce"] = "jovayeth:usdce";
    // Jovayeth testnet tokens
    UnderlyingAsset["tjovayeth:tcmn"] = "tjovayeth:tcmn";
    UnderlyingAsset["tjovayeth:usdce"] = "tjovayeth:usdce";
    // X Layer (OKB) mainnet tokens
    UnderlyingAsset["okbxlayer:usdg"] = "okbxlayer:usdg";
    UnderlyingAsset["okbxlayer:usdt0"] = "okbxlayer:usdt0";
    UnderlyingAsset["okbxlayer:usdt"] = "okbxlayer:usdt";
    UnderlyingAsset["okbxlayer:usdc"] = "okbxlayer:usdc";
    // X Layer (OKB) testnet tokens
    UnderlyingAsset["tokbxlayer:tzeb"] = "tokbxlayer:tzeb";
    // Story testnet tokens
    UnderlyingAsset["tip:usdc"] = "tip:usdc";
    UnderlyingAsset["tip:tmt"] = "tip:tmt";
    // Story mainnet tokens
    UnderlyingAsset["ip:aria"] = "ip:aria";
    // MegaEth mainnet tokens
    UnderlyingAsset["megaeth:mega"] = "megaeth:mega";
    UnderlyingAsset["megaeth:weth"] = "megaeth:weth";
    // MegaEth testnet tokens
    UnderlyingAsset["tmegaeth:tmt"] = "tmegaeth:tmt";
    // Plume testnet tokens
    UnderlyingAsset["tplume:usdc"] = "tplume:usdc";
    // Arbitrum testnet tokens
    UnderlyingAsset["tarbeth:link"] = "tarbeth:link";
    UnderlyingAsset["tarbeth:xsgd"] = "tarbeth:xsgd";
    // Optimism mainnet tokens
    UnderlyingAsset["opeth:link"] = "opeth:link";
    UnderlyingAsset["opeth:usdc"] = "opeth:usdc";
    UnderlyingAsset["opeth:usdcv2"] = "opeth:usdcv2";
    UnderlyingAsset["opeth:usdt"] = "opeth:usdt";
    UnderlyingAsset["opeth:op"] = "opeth:op";
    UnderlyingAsset["opeth:exa"] = "opeth:exa";
    UnderlyingAsset["opeth:wld"] = "opeth:wld";
    UnderlyingAsset["opeth:wct"] = "opeth:wct";
    UnderlyingAsset["opeth:spxux"] = "opeth:spxux";
    UnderlyingAsset["opeth:perp"] = "opeth:perp";
    UnderlyingAsset["opeth:flttx"] = "opeth:flttx";
    UnderlyingAsset["opeth:wtsix"] = "opeth:wtsix";
    UnderlyingAsset["opeth:modrx"] = "opeth:modrx";
    UnderlyingAsset["opeth:techx"] = "opeth:techx";
    UnderlyingAsset["opeth:wtsyx"] = "opeth:wtsyx";
    UnderlyingAsset["opeth:wtlgx"] = "opeth:wtlgx";
    UnderlyingAsset["opeth:wttsx"] = "opeth:wttsx";
    UnderlyingAsset["opeth:tipsx"] = "opeth:tipsx";
    UnderlyingAsset["opeth:wtstx"] = "opeth:wtstx";
    UnderlyingAsset["opeth:wtgxx"] = "opeth:wtgxx";
    UnderlyingAsset["opeth:lngvx"] = "opeth:lngvx";
    UnderlyingAsset["opeth:eqtyx"] = "opeth:eqtyx";
    UnderlyingAsset["opeth:velo"] = "opeth:velo";
    // Optimism testnet tokens
    UnderlyingAsset["topeth:terc18dp"] = "topeth:terc18dp";
    UnderlyingAsset["topeth:wct"] = "topeth:wct";
    // zkSync mainnet tokens
    UnderlyingAsset["zketh:link"] = "zketh:link";
    // zkSync testnet tokens
    UnderlyingAsset["tzketh:link"] = "tzketh:link";
    // Celo mainnet tokens
    UnderlyingAsset["celo:pact"] = "celo:pact";
    // bera mainnet tokens
    UnderlyingAsset["bera:bgt"] = "bera:bgt";
    UnderlyingAsset["bera:honey"] = "bera:honey";
    UnderlyingAsset["bera:usdc"] = "bera:usdc";
    UnderlyingAsset["bera:ibera"] = "bera:ibera";
    UnderlyingAsset["bera:dolo"] = "bera:dolo";
    UnderlyingAsset["bera:wgbera"] = "bera:wgbera";
    UnderlyingAsset["bera:ir"] = "bera:ir";
    // bera testnet tokens
    UnderlyingAsset["tbera:bgt"] = "tbera:bgt";
    UnderlyingAsset["tbera:honey"] = "tbera:honey";
    UnderlyingAsset["tbera:usdc"] = "tbera:usdc";
    UnderlyingAsset["tbera:ibera"] = "tbera:ibera";
    // Soneium NFTs
    // generic NFTs
    UnderlyingAsset["erc721:soneiumtoken"] = "erc721:soneiumtoken";
    UnderlyingAsset["erc1155:soneiumtoken"] = "erc1155:soneiumtoken";
    UnderlyingAsset["tsoneium:test721"] = "tsoneium:test721";
    UnderlyingAsset["tsoneium:test1155"] = "tsoneium:test1155";
    // coredao mainnet tokens
    UnderlyingAsset["coredao:stcore"] = "coredao:stcore";
    // coredao testnet tokens
    UnderlyingAsset["tcoredao:stcore"] = "tcoredao:stcore";
    UnderlyingAsset["tcoredao:tlstbtc"] = "tcoredao:tlstbtc";
    // Chiliz mainnet tokens
    UnderlyingAsset["chiliz:acm"] = "chiliz:acm";
    UnderlyingAsset["chiliz:afc"] = "chiliz:afc";
    UnderlyingAsset["chiliz:arg"] = "chiliz:arg";
    UnderlyingAsset["chiliz:asr"] = "chiliz:asr";
    UnderlyingAsset["chiliz:atm"] = "chiliz:atm";
    UnderlyingAsset["chiliz:avl"] = "chiliz:avl";
    UnderlyingAsset["chiliz:bar"] = "chiliz:bar";
    UnderlyingAsset["chiliz:benfica"] = "chiliz:benfica";
    UnderlyingAsset["chiliz:city"] = "chiliz:city";
    UnderlyingAsset["chiliz:gal"] = "chiliz:gal";
    UnderlyingAsset["chiliz:inter"] = "chiliz:inter";
    UnderlyingAsset["chiliz:ita"] = "chiliz:ita";
    UnderlyingAsset["chiliz:juv"] = "chiliz:juv";
    UnderlyingAsset["chiliz:mengo"] = "chiliz:mengo";
    UnderlyingAsset["chiliz:nap"] = "chiliz:nap";
    UnderlyingAsset["chiliz:og"] = "chiliz:og";
    UnderlyingAsset["chiliz:pepper"] = "chiliz:pepper";
    UnderlyingAsset["chiliz:por"] = "chiliz:por";
    UnderlyingAsset["chiliz:psg"] = "chiliz:psg";
    UnderlyingAsset["chiliz:spurs"] = "chiliz:spurs";
    UnderlyingAsset["chiliz:tra"] = "chiliz:tra";
    //world chain mainnet tokens
    UnderlyingAsset["world:wld"] = "world:wld";
    UnderlyingAsset["world:usdc"] = "world:usdc";
    //world chain testnet tokens
    UnderlyingAsset["tworld:wld"] = "tworld:wld";
    UnderlyingAsset["tworld:usdc"] = "tworld:usdc";
    // Flr mainnet tokens
    UnderlyingAsset["flr:wflr"] = "flr:wflr";
    // Flr testnet tokens
    UnderlyingAsset["tflr:wflr"] = "tflr:wflr";
    UnderlyingAsset["tflr:twc2flr"] = "tflr:twc2flr";
    UnderlyingAsset["ERC721"] = "erc721";
    UnderlyingAsset["ERC1155"] = "erc1155";
    UnderlyingAsset["NONSTANDARD"] = "nonstandard";
    // Cardano Token
    UnderlyingAsset["adaTestnetToken"] = "temporary-placeholder";
    // solana token
    UnderlyingAsset["sol:bome"] = "sol:bome";
    UnderlyingAsset["3uejh-usdc"] = "3uejh-usdc";
    UnderlyingAsset["avax-usdc"] = "avax-usdc";
    UnderlyingAsset["bop-usdc"] = "bop-usdc";
    UnderlyingAsset["sol:crown"] = "sol:crown";
    UnderlyingAsset["elu-usdt"] = "elu-usdt";
    UnderlyingAsset["fida-usdc"] = "fida-usdc";
    UnderlyingAsset["fida-usdt"] = "fida-usdt";
    UnderlyingAsset["ftt-ftt"] = "ftt-ftt";
    UnderlyingAsset["link-usdc"] = "link-usdc";
    UnderlyingAsset["lqid-usdc"] = "lqid-usdc";
    UnderlyingAsset["maticpo-usdc"] = "maticpo-usdc";
    UnderlyingAsset["msol-sol"] = "msol-sol";
    UnderlyingAsset["msol-usdc"] = "msol-usdc";
    UnderlyingAsset["prism-usdc"] = "prism-usdc";
    UnderlyingAsset["sol:pyth"] = "sol:pyth";
    UnderlyingAsset["rendoge-usdc"] = "rendoge-usdc";
    UnderlyingAsset["shdw-usdc"] = "shdw-usdc";
    UnderlyingAsset["sol-wtust"] = "sol-wtust";
    UnderlyingAsset["srm-usdc"] = "srm-usdc";
    UnderlyingAsset["srmet-srm"] = "srmet-srm";
    UnderlyingAsset["sushi-usdc"] = "sushi-usdc";
    UnderlyingAsset["tuatlas"] = "tuatlas";
    UnderlyingAsset["tucope"] = "tucope";
    UnderlyingAsset["tulike"] = "tulike";
    UnderlyingAsset["tureal"] = "tureal";
    UnderlyingAsset["tusamo"] = "tusamo";
    UnderlyingAsset["usdt-usdc"] = "usdt-usdc";
    UnderlyingAsset["wbwbnb-usdc"] = "wbwbnb-usdc";
    UnderlyingAsset["wheth-usdc"] = "wheth-usdc";
    UnderlyingAsset["wtust-usdt"] = "wtust-usdt";
    UnderlyingAsset["xcope-usdc"] = "xcope-usdc";
    UnderlyingAsset["xrp-sollet"] = "xrp-sollet";
    UnderlyingAsset["aury"] = "aury";
    UnderlyingAsset["dio"] = "dio";
    UnderlyingAsset["sol-perp"] = "sol-perp";
    UnderlyingAsset["sol-woo"] = "sol-woo";
    UnderlyingAsset["sol-weth"] = "sol-weth";
    UnderlyingAsset["btc-sollet"] = "btc-sollet";
    UnderlyingAsset["eth-sollet"] = "eth-sollet";
    UnderlyingAsset["sol:bonk"] = "sol:bonk";
    UnderlyingAsset["jto"] = "jto";
    UnderlyingAsset["sol:jto"] = "sol:jto";
    UnderlyingAsset["jup"] = "jup";
    UnderlyingAsset["sol:jup"] = "sol:jup";
    UnderlyingAsset["sol:honey"] = "sol:honey";
    UnderlyingAsset["mobile"] = "mobile";
    UnderlyingAsset["sol:mobile"] = "sol:mobile";
    UnderlyingAsset["wif"] = "wif";
    UnderlyingAsset["sol:wif"] = "sol:wif";
    UnderlyingAsset["natix"] = "natix";
    UnderlyingAsset["sol:natix"] = "sol:natix";
    UnderlyingAsset["sol:ks"] = "sol:ks";
    UnderlyingAsset["sol:apusdt"] = "sol:apusdt";
    UnderlyingAsset["sol:acusd"] = "sol:acusd";
    UnderlyingAsset["sol:solink"] = "sol:solink";
    UnderlyingAsset["sol:soon"] = "sol:soon";
    UnderlyingAsset["sol:wylds"] = "sol:wylds";
    UnderlyingAsset["sol:block"] = "sol:block";
    UnderlyingAsset["sol:render"] = "sol:render";
    UnderlyingAsset["sol:wen"] = "sol:wen";
    UnderlyingAsset["sol:mew"] = "sol:mew";
    UnderlyingAsset["sol:pyusd"] = "sol:pyusd";
    UnderlyingAsset["sol:moveusd"] = "sol:moveusd";
    UnderlyingAsset["sol:dxl"] = "sol:dxl";
    UnderlyingAsset["sol:mother"] = "sol:mother";
    UnderlyingAsset["sol:wrose"] = "sol:wrose";
    UnderlyingAsset["sol:atlas"] = "sol:atlas";
    UnderlyingAsset["sol:mdt"] = "sol:mdt";
    UnderlyingAsset["sol:io"] = "sol:io";
    UnderlyingAsset["sol:aave"] = "sol:aave";
    UnderlyingAsset["sol:ldo"] = "sol:ldo";
    UnderlyingAsset["sol:gt"] = "sol:gt";
    UnderlyingAsset["sol:popcat"] = "sol:popcat";
    UnderlyingAsset["sol:axs"] = "sol:axs";
    UnderlyingAsset["sol:sand"] = "sol:sand";
    UnderlyingAsset["sol:ens"] = "sol:ens";
    UnderlyingAsset["sol:enron"] = "sol:enron";
    UnderlyingAsset["sol:jitosol"] = "sol:jitosol";
    UnderlyingAsset["sol:zeus"] = "sol:zeus";
    UnderlyingAsset["sol:kmno"] = "sol:kmno";
    UnderlyingAsset["sol:giga"] = "sol:giga";
    UnderlyingAsset["sol:tnsr"] = "sol:tnsr";
    UnderlyingAsset["sol:ssol"] = "sol:ssol";
    UnderlyingAsset["sol:drift"] = "sol:drift";
    UnderlyingAsset["sol:spx"] = "sol:spx";
    UnderlyingAsset["sol:turbo"] = "sol:turbo";
    UnderlyingAsset["sol:fartcoin"] = "sol:fartcoin";
    UnderlyingAsset["sol:swarms"] = "sol:swarms";
    UnderlyingAsset["sol:nc"] = "sol:nc";
    UnderlyingAsset["sol:tai"] = "sol:tai";
    UnderlyingAsset["sol:pengu"] = "sol:pengu";
    UnderlyingAsset["sol:corn"] = "sol:corn";
    UnderlyingAsset["sol:yes"] = "sol:yes";
    UnderlyingAsset["sol:ai16z"] = "sol:ai16z";
    UnderlyingAsset["sol:pnut"] = "sol:pnut";
    UnderlyingAsset["sol:nyan"] = "sol:nyan";
    UnderlyingAsset["sol:fight"] = "sol:fight";
    UnderlyingAsset["sol:wet"] = "sol:wet";
    UnderlyingAsset["sol:meta"] = "sol:meta";
    UnderlyingAsset["sol:portals"] = "sol:portals";
    UnderlyingAsset["sol:virtual"] = "sol:virtual";
    UnderlyingAsset["sol:zerebro"] = "sol:zerebro";
    UnderlyingAsset["sol:arc"] = "sol:arc";
    UnderlyingAsset["sol:nos"] = "sol:nos";
    UnderlyingAsset["sol:jlp"] = "sol:jlp";
    UnderlyingAsset["sol:grass"] = "sol:grass";
    UnderlyingAsset["sol:trump"] = "sol:trump";
    UnderlyingAsset["sol:melania"] = "sol:melania";
    UnderlyingAsset["sol:ustry"] = "sol:ustry";
    UnderlyingAsset["sol:eurob"] = "sol:eurob";
    UnderlyingAsset["sol:tesouro"] = "sol:tesouro";
    UnderlyingAsset["sol:cetes"] = "sol:cetes";
    UnderlyingAsset["sol:gilts"] = "sol:gilts";
    UnderlyingAsset["sol:muskit"] = "sol:muskit";
    UnderlyingAsset["sol:matrix"] = "sol:matrix";
    UnderlyingAsset["sol:eurcv"] = "sol:eurcv";
    UnderlyingAsset["sol:layer"] = "sol:layer";
    UnderlyingAsset["sol:rock"] = "sol:rock";
    UnderlyingAsset["sol:dood"] = "sol:dood";
    UnderlyingAsset["sol:sb"] = "sol:sb";
    UnderlyingAsset["sol:dfdvsol"] = "sol:dfdvsol";
    UnderlyingAsset["sol:chillguy"] = "sol:chillguy";
    UnderlyingAsset["sol:moodeng"] = "sol:moodeng";
    UnderlyingAsset["sol:hsol"] = "sol:hsol";
    UnderlyingAsset["sol:grph"] = "sol:grph";
    UnderlyingAsset["sol:superbonds"] = "sol:superbonds";
    UnderlyingAsset["sol:would"] = "sol:would";
    UnderlyingAsset["sol:dog"] = "sol:dog";
    UnderlyingAsset["sol:dont"] = "sol:dont";
    UnderlyingAsset["sol:saros"] = "sol:saros";
    UnderlyingAsset["sol:babydoge"] = "sol:babydoge";
    UnderlyingAsset["sol:useless"] = "sol:useless";
    UnderlyingAsset["sol:gohome"] = "sol:gohome";
    UnderlyingAsset["sol:aura"] = "sol:aura";
    UnderlyingAsset["sol:me"] = "sol:me";
    UnderlyingAsset["sol:alch"] = "sol:alch";
    UnderlyingAsset["sol:launchcoin"] = "sol:launchcoin";
    UnderlyingAsset["sol:stik"] = "sol:stik";
    UnderlyingAsset["sol:chill"] = "sol:chill";
    UnderlyingAsset["sol:zbcn"] = "sol:zbcn";
    UnderlyingAsset["sol:benji"] = "sol:benji";
    UnderlyingAsset["sol:dupe"] = "sol:dupe";
    UnderlyingAsset["sol:tank"] = "sol:tank";
    UnderlyingAsset["sol:grift"] = "sol:grift";
    UnderlyingAsset["sol:usdk"] = "sol:usdk";
    UnderlyingAsset["sol:usdky"] = "sol:usdky";
    UnderlyingAsset["sol:wave"] = "sol:wave";
    UnderlyingAsset["sol:usdcv"] = "sol:usdcv";
    UnderlyingAsset["sol:2z"] = "sol:2z";
    UnderlyingAsset["sol:cloud"] = "sol:cloud";
    UnderlyingAsset["sol:eliza"] = "sol:eliza";
    UnderlyingAsset["sol:eurc"] = "sol:eurc";
    UnderlyingAsset["sol:dynosol"] = "sol:dynosol";
    UnderlyingAsset["sol:cipher"] = "sol:cipher";
    UnderlyingAsset["sol:bio"] = "sol:bio";
    UnderlyingAsset["sol:rekt"] = "sol:rekt";
    UnderlyingAsset["sol:xyo"] = "sol:xyo";
    UnderlyingAsset["sol:zig"] = "sol:zig";
    UnderlyingAsset["sol:xsgd"] = "sol:xsgd";
    UnderlyingAsset["sol:straitxusd"] = "sol:straitxusd";
    UnderlyingAsset["sol:usx"] = "sol:usx";
    UnderlyingAsset["sol:rksol"] = "sol:rksol";
    UnderlyingAsset["sol:usda1"] = "sol:usda1";
    UnderlyingAsset["sol:ponke"] = "sol:ponke";
    UnderlyingAsset["sol:jellyjelly"] = "sol:jellyjelly";
    UnderlyingAsset["sol:yu"] = "sol:yu";
    UnderlyingAsset["sol:met"] = "sol:met";
    UnderlyingAsset["sol:white"] = "sol:white";
    UnderlyingAsset["sol:yzy"] = "sol:yzy";
    UnderlyingAsset["sol:wild"] = "sol:wild";
    UnderlyingAsset["sol:npc"] = "sol:npc";
    UnderlyingAsset["sol:aix"] = "sol:aix";
    UnderlyingAsset["sol:troll"] = "sol:troll";
    UnderlyingAsset["sol:ban"] = "sol:ban";
    UnderlyingAsset["sol:dbr"] = "sol:dbr";
    UnderlyingAsset["sol:bless"] = "sol:bless";
    UnderlyingAsset["sol:cpool"] = "sol:cpool";
    UnderlyingAsset["sol:home"] = "sol:home";
    UnderlyingAsset["sol:oob"] = "sol:oob";
    UnderlyingAsset["sol:xnet"] = "sol:xnet";
    UnderlyingAsset["sol:xnetv2"] = "sol:xnetv2";
    UnderlyingAsset["sol:prcl"] = "sol:prcl";
    UnderlyingAsset["sol:asp"] = "sol:asp";
    UnderlyingAsset["tsol:txsgd"] = "sol:txsgd";
    UnderlyingAsset["tsol:txusd"] = "sol:txusd";
    // TRX tokens
    UnderlyingAsset["trx:htx"] = "trx:htx";
    UnderlyingAsset["trx:jst"] = "trx:jst";
    UnderlyingAsset["trx:tusd"] = "trx:tusd";
    UnderlyingAsset["trx:win"] = "trx:win";
    UnderlyingAsset["trx:btt"] = "trx:btt";
    UnderlyingAsset["trx:usdd"] = "trx:usdd";
    UnderlyingAsset["trx:usdt"] = "trx:usdt";
    UnderlyingAsset["trx:usd1"] = "trx:usd1";
    UnderlyingAsset["trx:nft"] = "trx:nft";
    UnderlyingAsset["trx:trxs"] = "trx:trxs";
    // TRX testnet tokens
    UnderlyingAsset["ttrx:usdt"] = "ttrx:usdt";
    UnderlyingAsset["ttrx:usd1"] = "ttrx:usd1";
    UnderlyingAsset["ttrx:stgusd1"] = "ttrx:stgusd1";
    // XRP tokens
    UnderlyingAsset["txrp:tst-rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd"] = "txrp:tst-rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd";
    UnderlyingAsset["xrp:rlusd"] = "xrp:rlusd";
    UnderlyingAsset["txrp:rlusd"] = "txrp:rlusd";
    UnderlyingAsset["txrp:xat"] = "txrp:xat";
    UnderlyingAsset["xrp:tbill"] = "xrp:tbill";
    UnderlyingAsset["xrp:xsgd"] = "xrp:xsgd";
    UnderlyingAsset["xrp:veur"] = "xrp:veur";
    UnderlyingAsset["xrp:vchf"] = "xrp:vchf";
    UnderlyingAsset["xrp:vgbp"] = "xrp:vgbp";
    UnderlyingAsset["xrp:solo"] = "xrp:solo";
    UnderlyingAsset["xrp:usdc"] = "xrp:usdc";
    UnderlyingAsset["xrp:eurcv"] = "xrp:eurcv";
    UnderlyingAsset["xrp:aau"] = "xrp:aau";
    UnderlyingAsset["xrp:fiuaxrp"] = "xrp:fiuaxrp";
    // XRP testnet tokens
    UnderlyingAsset["txrp:xsgd"] = "txrp:xsgd";
    // Sui tokens
    UnderlyingAsset["sui:deep"] = "sui:deep";
    UnderlyingAsset["sui:suins"] = "sui:suins";
    UnderlyingAsset["sui:suiusde"] = "sui:suiusde";
    UnderlyingAsset["sui:fdusd"] = "sui:fdusd";
    UnderlyingAsset["sui:usdc"] = "sui:usdc";
    UnderlyingAsset["sui:wusdc"] = "sui:wusdc";
    UnderlyingAsset["sui:sca"] = "sui:sca";
    UnderlyingAsset["sui:times"] = "sui:times";
    UnderlyingAsset["sui:fud"] = "sui:fud";
    UnderlyingAsset["sui:afsui"] = "sui:afsui";
    UnderlyingAsset["sui:navx"] = "sui:navx";
    UnderlyingAsset["sui:vsui"] = "sui:vsui";
    UnderlyingAsset["sui:send"] = "sui:send";
    UnderlyingAsset["sui:cetus"] = "sui:cetus";
    UnderlyingAsset["sui:wal"] = "sui:wal";
    UnderlyingAsset["sui:xmn"] = "sui:xmn";
    UnderlyingAsset["sui:xaum"] = "sui:xaum";
    UnderlyingAsset["sui:alkimi"] = "sui:alkimi";
    UnderlyingAsset["sui:dmc"] = "sui:dmc";
    UnderlyingAsset["sui:mmt"] = "sui:mmt";
    // Sui testnet tokens
    UnderlyingAsset["tsui:deep"] = "tsui:deep";
    UnderlyingAsset["tsui:wal"] = "tsui:wal";
    // Apt tokens
    UnderlyingAsset["apt:usd1"] = "apt:usd1";
    UnderlyingAsset["apt:usdt"] = "apt:usdt";
    UnderlyingAsset["apt:usdc"] = "apt:usdc";
    UnderlyingAsset["apt:pact"] = "apt:pact";
    UnderlyingAsset["apt:benji"] = "apt:benji";
    UnderlyingAsset["apt:lsd"] = "apt:lsd";
    UnderlyingAsset["apt:kgen"] = "apt:kgen";
    UnderlyingAsset["apt:scf1"] = "apt:scf1";
    // Apt mainnet NFT collections
    UnderlyingAsset["apt:h00ts"] = "apt:h00ts";
    // Apt testnet tokens
    UnderlyingAsset["tapt:stgusd1"] = "tapt:stgusd1";
    UnderlyingAsset["tapt:usd1"] = "tapt:usd1";
    UnderlyingAsset["tapt:usdt"] = "tapt:usdt";
    // Apt testnet NFT collections
    UnderlyingAsset["tapt:nftcollection1"] = "tapt:nftcollection1";
    UnderlyingAsset["tapt:beta3loanbook"] = "tapt:beta3loanbook";
    // Sip10 tokens
    UnderlyingAsset["stx:sbtc"] = "stx:sbtc";
    UnderlyingAsset["stx:ststx"] = "stx:ststx";
    UnderlyingAsset["stx:alex"] = "stx:alex";
    UnderlyingAsset["stx:aeusdc"] = "stx:aeusdc";
    UnderlyingAsset["stx:usdh"] = "stx:usdh";
    UnderlyingAsset["stx:susdh"] = "stx:susdh";
    UnderlyingAsset["stx:welsh"] = "stx:welsh";
    // Sip10 testnet tokens
    UnderlyingAsset["tstx:tsbtc"] = "tstx:tsbtc";
    UnderlyingAsset["tstx:tsip6dp"] = "tstx:tsip6dp";
    UnderlyingAsset["tstx:tsip8dp"] = "tstx:tsip8dp";
    // TAO testnet tokens
    UnderlyingAsset["ttao:apex"] = "ttao:apex";
    UnderlyingAsset["ttao:onion"] = "ttao:onion";
    UnderlyingAsset["ttao:templar"] = "ttao:templar";
    UnderlyingAsset["ttao:targon"] = "ttao:targon";
    // Polymesh testnet tokens
    UnderlyingAsset["tpolyx:nvbitgot"] = "tpolyx:nvbitgot";
    UnderlyingAsset["tpolyx:RAND176TM"] = "tpolyx:RAND176TM";
    UnderlyingAsset["tpolyx:WEBINRASSET3"] = "tpolyx:WEBINRASSET3";
    UnderlyingAsset["tpolyx:WEBINRASSET4"] = "tpolyx:WEBINRASSET4";
    UnderlyingAsset["tpolyx:WEBINRASSET5"] = "tpolyx:WEBINRASSET5";
    UnderlyingAsset["tpolyx:WEBINRASSET6"] = "tpolyx:WEBINRASSET6";
    UnderlyingAsset["tpolyx:WEBINRASSET7"] = "tpolyx:WEBINRASSET7";
    UnderlyingAsset["tpolyx:BULLRWA"] = "tpolyx:BULLRWA";
    UnderlyingAsset["tpolyx:0x4a002922d38b8a7f87484b9c65a7ca0c"] = "tpolyx:0x4a002922d38b8a7f87484b9c65a7ca0c";
    UnderlyingAsset["tpolyx:0x753b5a2cb97c83e8b1bd66ed6643e4e9"] = "tpolyx:0x753b5a2cb97c83e8b1bd66ed6643e4e9";
    // Polymesh mainnet tokens
    UnderlyingAsset["polyx:0xa0ce6bc4c60981e08eca6504656c99e6"] = "polyx:0xa0ce6bc4c60981e08eca6504656c99e6";
    UnderlyingAsset["polyx:TEST-TOKEN-1"] = "polyx:TEST-TOKEN-1";
    // Hbar tokens
    UnderlyingAsset["hbar:karate"] = "hbar:karate";
    UnderlyingAsset["hbar:sauce"] = "hbar:sauce";
    UnderlyingAsset["hbar:dovu"] = "hbar:dovu";
    UnderlyingAsset["hbar:pack"] = "hbar:pack";
    UnderlyingAsset["hbar:jam"] = "hbar:jam";
    UnderlyingAsset["hbar:berry"] = "hbar:berry";
    UnderlyingAsset["hbar:bonzo"] = "hbar:bonzo";
    UnderlyingAsset["hbar:co2e"] = "hbar:co2e";
    UnderlyingAsset["hbar:hsuite"] = "hbar:hsuite";
    UnderlyingAsset["hbar:usdc"] = "hbar:usdc";
    // Hbar Testnet tokens
    UnderlyingAsset["thbar:txsgd"] = "thbar:txsgd";
    // Hbarevm TEstnet Tokens
    UnderlyingAsset["thbarevm:hbarnativetoken"] = "thbarevm:hbarnativetoken";
    UnderlyingAsset["thbarevm:tmt"] = "thbarevm:tmt";
    UnderlyingAsset["thbarevm:tnt"] = "thbarevm:tnt";
    // Nep141 tokens
    UnderlyingAsset["near:usdc"] = "near:usdc";
    UnderlyingAsset["near:usdt"] = "near:usdt";
    UnderlyingAsset["near:mpdao"] = "near:mpdao";
    UnderlyingAsset["near:stnear"] = "near:stnear";
    // Nep141 testnet tokens
    UnderlyingAsset["tnear:tnep24dp"] = "tnear:tnep24dp";
    UnderlyingAsset["tnear:usdc"] = "tnear:usdc";
    UnderlyingAsset["tnear:stnear"] = "tnear:stnear";
    // VET tokens
    UnderlyingAsset["vet:vtho"] = "vet:vtho";
    // VET testnet tokens
    UnderlyingAsset["tvet:vtho"] = "tvet:vtho";
    // Vet mainnet NFTs
    UnderlyingAsset["vet:sdt"] = "vet:sdt";
    // Vet testnet NFTs
    UnderlyingAsset["tvet:sdt"] = "tvet:sdt";
    // COSMOS tokens
    UnderlyingAsset["hash:ylds"] = "hash:ylds";
    UnderlyingAsset["hash:figr"] = "hash:figr";
    // COSMOS testnet tokens
    UnderlyingAsset["thash:ylds"] = "thash:ylds";
    UnderlyingAsset["thash:tfigr"] = "thash:tfigr";
    // TON mainnet tokens
    UnderlyingAsset["ton:usdt"] = "ton:usdt";
    UnderlyingAsset["ton:usde"] = "ton:usde";
    UnderlyingAsset["ton:not"] = "ton:not";
    UnderlyingAsset["ton:cati"] = "ton:cati";
    UnderlyingAsset["ton:dogs"] = "ton:dogs";
    UnderlyingAsset["ton:ston"] = "ton:ston";
    // TON testnet tokens
    UnderlyingAsset["tton:ukwny-us"] = "tton:ukwny-us";
    UnderlyingAsset["eth:0x0"] = "eth:0x0";
    UnderlyingAsset["eth:vvs"] = "eth:vvs";
    UnderlyingAsset["eth:bmx"] = "eth:bmx";
    UnderlyingAsset["eth:pro"] = "eth:pro";
    UnderlyingAsset["eth:prime"] = "eth:prime";
    UnderlyingAsset["eth:pokt"] = "eth:pokt";
    UnderlyingAsset["eth:lon"] = "eth:lon";
    UnderlyingAsset["eth:rlb"] = "eth:rlb";
    UnderlyingAsset["eth:neiro2"] = "eth:neiro2";
    UnderlyingAsset["eth:sign"] = "eth:sign";
    UnderlyingAsset["eth:vsn"] = "eth:vsn";
    UnderlyingAsset["eth:shx"] = "eth:shx";
    UnderlyingAsset["eth:slay"] = "eth:slay";
    UnderlyingAsset["eth:mxnb"] = "eth:mxnb";
    UnderlyingAsset["eth:hwhlp"] = "eth:hwhlp";
    UnderlyingAsset["eth:mxnd"] = "eth:mxnd";
    UnderlyingAsset["eth:bio"] = "eth:bio";
    UnderlyingAsset["eth:prove"] = "eth:prove";
    UnderlyingAsset["eth:zrc"] = "eth:zrc";
    UnderlyingAsset["eth:open"] = "eth:open";
    UnderlyingAsset["eth:mbg"] = "eth:mbg";
    UnderlyingAsset["eth:rekt"] = "eth:rekt";
    UnderlyingAsset["eth:shvon"] = "eth:shvon";
    UnderlyingAsset["eth:pt-cusdo-28may2026"] = "eth:pt-cusdo-28may2026";
    UnderlyingAsset["eth:usd1cx"] = "eth:usd1cx";
    UnderlyingAsset["eth:ctusd1cx"] = "eth:ctusd1cx";
    // ADA testnet tokens
    UnderlyingAsset["tada:water"] = "tada:water";
    UnderlyingAsset["tada:tusda"] = "tada:tusda";
    // ADA mainnet tokens
    UnderlyingAsset["ada:min"] = "ada:min";
    UnderlyingAsset["ada:snek"] = "ada:snek";
    UnderlyingAsset["ada:wmtx"] = "ada:wmtx";
    UnderlyingAsset["ada:iag"] = "ada:iag";
    UnderlyingAsset["ada:djed"] = "ada:djed";
    UnderlyingAsset["ada:usda"] = "ada:usda";
    UnderlyingAsset["ada:night"] = "ada:night";
    UnderlyingAsset["ada:lcc"] = "ada:lcc";
    UnderlyingAsset["ada:awlf"] = "ada:awlf";
    UnderlyingAsset["ada:asnek"] = "ada:asnek";
    // Canton testnet tokens
    UnderlyingAsset["tcanton:testcoin1"] = "tcanton:testcoin1";
    UnderlyingAsset["tcanton:testtoken"] = "tcanton:testtoken";
    // Canton mainnet tokens
    UnderlyingAsset["canton:usdcx"] = "canton:usdcx";
    UnderlyingAsset["canton:cbtc"] = "canton:cbtc";
    UnderlyingAsset["canton:usdxlr"] = "canton:usdxlr";
    // Tempo mainnet tokens
    UnderlyingAsset["tempo:pathusd"] = "tempo:pathusd";
    UnderlyingAsset["tempo:usdc"] = "tempo:usdc";
    // Tempo testnet tokens
    UnderlyingAsset["ttempo:pathusd"] = "ttempo:pathusd";
    UnderlyingAsset["ttempo:alphausd"] = "ttempo:alphausd";
    UnderlyingAsset["ttempo:betausd"] = "ttempo:betausd";
    UnderlyingAsset["ttempo:thetausd"] = "ttempo:thetausd";
    // fiats
    UnderlyingAsset["AED"] = "aed";
    UnderlyingAsset["EUR"] = "eur";
    UnderlyingAsset["GBP"] = "gbp";
    UnderlyingAsset["INR"] = "inr";
    UnderlyingAsset["SGD"] = "sgd";
    UnderlyingAsset["USD"] = "usd";
})(UnderlyingAsset || (exports.UnderlyingAsset = UnderlyingAsset = {}));
/**
 * This is the curve BitGo signs against with the user, backup and BitGo key.
 */
var KeyCurve;
(function (KeyCurve) {
    KeyCurve["Secp256k1"] = "secp256k1";
    KeyCurve["Ed25519"] = "ed25519";
})(KeyCurve || (exports.KeyCurve = KeyCurve = {}));
/**
 * This enum contains the base units for the coins that BitGo supports
 */
var BaseUnit;
(function (BaseUnit) {
    BaseUnit["ATOM"] = "uatom";
    BaseUnit["APT"] = "octa";
    BaseUnit["ETH"] = "wei";
    BaseUnit["BABY"] = "ubbn";
    BaseUnit["BTC"] = "satoshi";
    BaseUnit["BSC"] = "jager";
    BaseUnit["XLM"] = "stroop";
    BaseUnit["TRX"] = "sun";
    BaseUnit["HBAR"] = "tinybar";
    BaseUnit["ALGO"] = "microAlgo";
    BaseUnit["EOS"] = "eos";
    BaseUnit["SOL"] = "lamport";
    BaseUnit["ADA"] = "lovelace";
    BaseUnit["USD"] = "USD";
    BaseUnit["LNBTC"] = "millisatoshi";
    BaseUnit["LTC"] = "microlitecoins";
    BaseUnit["DASH"] = "duff";
    BaseUnit["ZEC"] = "zatoshi";
    BaseUnit["CSPR"] = "mote";
    BaseUnit["DOT"] = "planck";
    BaseUnit["XRP"] = "drop";
    BaseUnit["XTZ"] = "micro xtz";
    BaseUnit["STX"] = "micro-STX";
    BaseUnit["SUI"] = "MIST";
    BaseUnit["TON"] = "nanoton";
    BaseUnit["NEAR"] = "yocto";
    BaseUnit["OFC"] = "ofcCoin";
    BaseUnit["OSMO"] = "uosmo";
    BaseUnit["FIAT"] = "fiatCoin";
    BaseUnit["TIA"] = "utia";
    BaseUnit["HASH"] = "nhash";
    BaseUnit["BLD"] = "ubld";
    BaseUnit["SEI"] = "usei";
    BaseUnit["INJECTIVE"] = "inj";
    BaseUnit["IOTA"] = "iota";
    BaseUnit["ZETA"] = "azeta";
    BaseUnit["KAVACOSMOS"] = "ukava";
    BaseUnit["DYDX"] = "adydx";
    BaseUnit["COREUM"] = "ucore";
    BaseUnit["TCOREUM"] = "utestcore";
    BaseUnit["ISLM"] = "aISLM";
    BaseUnit["RUNE"] = "rune";
    BaseUnit["TAO"] = "rao";
    BaseUnit["ICP"] = "e8s";
    BaseUnit["MANTRA"] = "uom";
    BaseUnit["POLYX"] = "micropolyx";
    BaseUnit["CRONOS"] = "basecro";
    BaseUnit["FETCHAI"] = "afet";
    BaseUnit["INITIA"] = "uinit";
    BaseUnit["ASI"] = "afet";
    BaseUnit["VET"] = "wei";
    BaseUnit["TCRONOS"] = "basetcro";
    BaseUnit["TASI"] = "atestfet";
    BaseUnit["CANTON"] = "canton";
    BaseUnit["USDC"] = "usdc";
})(BaseUnit || (exports.BaseUnit = BaseUnit = {}));
class BaseCoin {
    /**
     * Ensures that the base coin constructor was passed a valid set of options.
     *
     * This includes checking that:
     * - All coin features of the new instance are allowed by the coin class
     * - No features required by the coin class are missing from the new instance
     * @param {BaseCoinConstructorOptions} options
     * @throws {DisallowedCoinFeatureError} if any of the coin features are not allowed for the coin class
     * @throws {MissingRequiredCoinFeatureError} if any features required by the coin class are missing
     */
    validateOptions(options) {
        const requiredFeatures = this.requiredFeatures();
        const disallowedFeatures = this.disallowedFeatures();
        const intersectionFeatures = Array.from(requiredFeatures).filter((feat) => disallowedFeatures.has(feat));
        if (intersectionFeatures.length > 0) {
            throw new errors_1.ConflictingCoinFeaturesError(options.name, intersectionFeatures);
        }
        for (const feature of options.features) {
            if (disallowedFeatures.has(feature)) {
                throw new errors_1.DisallowedCoinFeatureError(options.name, feature);
            }
            if (requiredFeatures.has(feature)) {
                requiredFeatures.delete(feature);
            }
        }
        if (requiredFeatures.size > 0) {
            // some required features were missing
            throw new errors_1.MissingRequiredCoinFeatureError(options.name, Array.from(requiredFeatures));
        }
        // assets require a valid uuid v4 id
        if (!BaseCoin.isValidUuidV4(options.id)) {
            throw new errors_1.InvalidIdError(options.name, options.id);
        }
    }
    constructor(options) {
        this.validateOptions(options);
        this.id = options.id;
        this.fullName = options.fullName;
        this.name = options.name;
        this.alias = options.alias;
        this.prefix = options.prefix;
        this.suffix = options.suffix;
        this.denom = options.denom;
        this.baseUnit = options.baseUnit;
        this.kind = options.kind;
        this.family = options.network.family;
        this.isToken = options.isToken;
        this.features = options.features;
        this.decimalPlaces = options.decimalPlaces;
        this.asset = options.asset;
        this.network = options.network;
        this.primaryKeyCurve = options.primaryKeyCurve;
    }
    /**
     * Returns features from a base feature set, excluding specified features
     * @param excludedFeatures Array of features to exclude
     * @param baseFeatures Base feature array to filter from (optional)
     * @returns Filtered array of features
     */
    static getFeaturesByTypeExcluding(excludedFeatures, baseFeatures) {
        if (!baseFeatures) {
            return [];
        }
        return baseFeatures.filter((feature) => !excludedFeatures.includes(feature));
    }
}
exports.BaseCoin = BaseCoin;
BaseCoin.isValidUuidV4 = (uuid) => {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
    return uuidRegex.test(uuid);
};
