// @bitgo/statics/dist/src/canton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Canton = void 0;
exports.canton = canton;
const base_1 = require("./base");
class Canton extends base_1.BaseCoin {
    constructor(options) {
        super({
            ...options,
            kind: base_1.CoinKind.CRYPTO,
            isToken: false,
            decimalPlaces: 10,
            baseUnit: base_1.BaseUnit.CANTON,
        });
        this.network = options.network;
    }
    disallowedFeatures() {
        return new Set([base_1.CoinFeature.ACCOUNT_MODEL]);
    }
    requiredFeatures() {
        return new Set([]);
    }
}
exports.Canton = Canton;
/**
 * Factory function for canton coin instances
 *
 * @param {String} id unique identifier (uuid v4)
 * @param {String} name unique identifier of the coin
 * @param {String} fullName complete human-readable name of the coin
 * @param {BaseNetwork} network network object for this coin
 * @param {UnderlyingAsset} asset asset which this coin represents. This is the same for both mainNet and testNet variants of a coin.
 * @param {string} prefix optional coin prefix. Defaults to empty string
 * @param {string} suffix optional coin suffix. Defaults to coin name.
 * @param {CoinFeature[]} features features of this coin. Defaults to the CANTON_DEFAULT_FEATURES defined in `account`
 * @param {KeyCurve} primaryKeyCurve the elliptic curve for this chain/token
 */
function canton(id, name, fullName, network, asset, features, primaryKeyCurve = base_1.KeyCurve.Ed25519, prefix = '', suffix = name.toUpperCase()) {
    return Object.freeze(new Canton({
        id,
        name,
        fullName,
        network,
        features,
        asset,
        prefix,
        suffix,
        primaryKeyCurve,
    }));
}
