// @bitgo/statics/dist/src/coins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.coins = void 0;
exports.createToken = createToken;
exports.isCoinPresentInCoinMap = isCoinPresentInCoinMap;
exports.createTokenMapUsingConfigDetails = createTokenMapUsingConfigDetails;
exports.createTokenMapUsingTrimmedConfigDetails = createTokenMapUsingTrimmedConfigDetails;
exports.createTokenUsingTrimmedConfigDetails = createTokenUsingTrimmedConfigDetails;
const account_1 = require("./account");
const ofc_1 = require("./ofc");
const base_1 = require("./base");
const map_1 = require("./map");
const networks_1 = require("./networks");
const networkFeatureMapForTokens_1 = require("./networkFeatureMapForTokens");
const ofcErc20Coins_1 = require("./coins/ofcErc20Coins");
const ofcCoins_1 = require("./coins/ofcCoins");
const allCoinsAndTokens_1 = require("./allCoinsAndTokens");
const botOfcTokens_1 = require("./coins/botOfcTokens");
exports.coins = map_1.CoinMap.fromCoins([
    ...allCoinsAndTokens_1.allCoinsAndTokens,
    ...ofcErc20Coins_1.ofcErc20Coins,
    ...ofcErc20Coins_1.tOfcErc20Coins,
    ...ofcCoins_1.ofcCoins,
    ...botOfcTokens_1.botOfcTokens,
]);
// Build a map of ERC20-supporting chain family names to their mainnet coin names
// Maps family -> coin name (e.g., 'ip' -> 'ip')
const erc20ChainToNameMap = {};
allCoinsAndTokens_1.allCoinsAndTokens.forEach((coin) => {
    if (coin.features.includes(base_1.CoinFeature.SUPPORTS_ERC20) &&
        coin.network.type === networks_1.NetworkType.MAINNET &&
        !coin.isToken) {
        erc20ChainToNameMap[coin.family] = coin.name;
    }
});
const erc721ChainToNameMap = {};
allCoinsAndTokens_1.allCoinsAndTokens.forEach((coin) => {
    if (coin.features.includes(base_1.CoinFeature.SUPPORTS_ERC721) &&
        coin.network.type === networks_1.NetworkType.MAINNET &&
        !coin.isToken) {
        erc721ChainToNameMap[coin.family] = coin.name;
    }
});
function createToken(token) {
    const initializerMap = {
        algo: account_1.algoToken,
        apt: account_1.aptToken,
        arbeth: account_1.arbethErc20,
        avaxc: account_1.avaxErc20,
        baseeth: account_1.erc20Token,
        bera: account_1.beraErc20,
        flow: account_1.erc20Token,
        lineaeth: account_1.erc20Token,
        seievm: account_1.erc20Token,
        mon: account_1.erc20Token,
        xdc: account_1.erc20Token,
        bsc: account_1.bscToken,
        celo: account_1.celoToken,
        cosmos: account_1.cosmosToken,
        eth: account_1.erc20,
        eos: account_1.eosToken,
        hbar: account_1.hederaToken,
        near: account_1.nep141Token,
        opeth: account_1.opethErc20,
        polygon: account_1.polygonErc20,
        sol: account_1.solToken,
        stx: account_1.sip10Token,
        sui: account_1.suiToken,
        tao: account_1.taoToken,
        polyx: account_1.polyxToken,
        trx: account_1.tronToken,
        vet: account_1.vetToken,
        xlm: account_1.stellarToken,
        xrp: account_1.xrpToken,
        ofc: ofc_1.ofcToken,
        ada: account_1.adaToken,
        ton: account_1.jettonToken,
    };
    // dynamically add erc20 token initializers for eth like chains to the initializer map
    Object.keys(erc20ChainToNameMap).forEach((key) => {
        initializerMap[key] = account_1.erc20Token;
    });
    Object.keys(erc721ChainToNameMap).forEach((key) => {
        initializerMap[key] = account_1.erc721Token;
    });
    //return the BaseCoin from default coin map if present
    if (isCoinPresentInCoinMap({ ...token })) {
        if (exports.coins.has(token.name)) {
            return exports.coins.get(token.name);
        }
        if (exports.coins.has(token.id)) {
            return exports.coins.get(token.id);
        }
        if (token.alias && exports.coins.has(token.alias)) {
            return exports.coins.get(token.alias);
        }
    }
    const family = token.family;
    const initializer = initializerMap[family];
    if (!initializer) {
        return undefined;
    }
    const commonArgs = [
        token.id,
        token.name,
        token.fullName,
        token.decimalPlaces,
        token.asset,
        token.features,
        token.prefix,
        token.suffix,
        token.network,
        token.primaryKeyCurve,
    ];
    switch (family) {
        case erc20ChainToNameMap[family]:
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.contractAddress || token.tokenAddress, // contractAddress
            token.asset, token.network, token.features, token.prefix, token.suffix, token.primaryKeyCurve);
        case erc721ChainToNameMap[family]:
            return initializer(...commonArgs.slice(0, 3), // id, name, fullName
            token.contractAddress, // contractAddress
            token.network, token.features, token.prefix, token.suffix, token.primaryKeyCurve);
        case 'arbeth':
        case 'avaxc':
        case 'baseeth':
        case 'bera':
        case 'bsc':
        case 'flow':
        case 'lineaeth':
        case 'seievm':
        case 'mon':
        case 'xdc':
        case 'celo':
        case 'eth':
        case 'opeth':
        case 'polygon':
        case 'trx':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.contractAddress || token.tokenAddress, // contractAddress
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'apt':
            const { initFunc, objectId } = getAptTokenInitializer(token);
            return initFunc(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            objectId, ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'vet':
            const { vetInitFunc, vetObjectId, isNFT } = getVetTokenInitializer(token);
            if (isNFT) {
                return vetInitFunc(...commonArgs.slice(0, 3), // id, name, fullName
                vetObjectId, ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
                );
            }
            else {
                return vetInitFunc(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
                vetObjectId, ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
                );
            }
        case 'stx':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.assetId, // assetId
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'algo':
            return initializer(...commonArgs.slice(0, 2), // id, name
            token.alias, // alias
            ...commonArgs.slice(2) // fullName, decimal, asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'eos':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.contractName, // contractName
            token.contractAddress, // contractAddress
            ...commonArgs.slice(4, 8), // asset, features, prefix, suffix
            token.symbol, // symbol
            ...commonArgs.slice(8) // network, primaryKeyCurve
            );
        case 'hbar':
            return initializer(...commonArgs.slice(0, 3), // id, name, fullName
            token.network, // network
            token.decimalPlaces, token.asset, token.contractAddress, // contractAddress
            ...commonArgs.slice(5, 8), // features, prefix, suffix
            token.primaryKeyCurve);
        case 'sol':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.tokenAddress, // tokenAddress
            token.contractAddress, // contractAddress
            token.asset, token.features, token.programId, ...commonArgs.slice(6) // prefix, suffix, network, primaryKeyCurve
            );
        case 'sui':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.packageId, // packageId
            token.module, // module
            token.symbol, // symbol
            token.contractAddress, // contractAddress
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'tao':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.subnetId, // subnetId
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'polyx':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.ticker, // ticker
            token.assetId, // assetId
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'xlm':
            return initializer(...commonArgs.slice(0, 5), // id, name, fullName, decimalPlaces, asset
            token.domain, // domain
            ...commonArgs.slice(5) // features, prefix, suffix, network, primaryKeyCurve
            );
        case 'xrp':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.issuerAddress, // issuerAddress
            token.currecnycode, // currencyCode
            token.contractAddress, // contractAddress
            token.domain, // domain
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'ofc':
            return initializer(...commonArgs, // id, name, fullName, decimalPlaces, asset, prefix, suffix, network, primaryKeyCurve
            token.baseUnit, // baseUnit
            token.isToken, // isToken
            token.kind, // kind
            token.addressCoin // addressCoin
            );
        case 'asi':
        case 'atom':
        case 'baby':
        case 'bld':
        case 'coreum':
        case 'cronos':
        case 'hash':
        case 'injective':
        case 'initia':
        case 'osmo':
        case 'thor':
        case 'sei':
        case 'tia':
        case 'zeta':
            return initializer(...commonArgs.slice(0, 3), // id, name, fullName
            token.denom, // denom
            token.decimalPlaces, // decimalPlaces
            token.network, // network
            token.baseUnit, // baseUnit
            ...commonArgs.slice(4, 8), // asset, features, prefix, suffix
            token.primaryKeyCurve // primaryKeyCurve
            );
        case 'ada':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.assetName, token.policyId, token.contractAddress, ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        case 'ton':
            return initializer(...commonArgs.slice(0, 4), // id, name, fullName, decimalPlaces
            token.contractAddress, // contractAddress
            ...commonArgs.slice(4) // asset, features, prefix, suffix, network, primaryKeyCurve
            );
        default:
            return undefined;
    }
}
function getAptTokenInitializer(token) {
    if (token.assetId) {
        // used for fungible-assets / legacy coins etc.
        return {
            initFunc: account_1.aptToken,
            objectId: token.assetId,
        };
    }
    // used for non-fungible token (NFT) collections
    return {
        initFunc: account_1.aptNFTCollection,
        objectId: token.nftCollectionId,
    };
}
function getVetTokenInitializer(token) {
    if (token.nftCollectionId) {
        return {
            vetInitFunc: account_1.vetNFTCollection,
            vetObjectId: token.nftCollectionId,
            isNFT: true,
        };
    }
    return {
        vetInitFunc: account_1.vetToken,
        vetObjectId: token.contractAddress,
        isNFT: false,
    };
}
function isCoinPresentInCoinMap({ name, id, alias }) {
    return Boolean(exports.coins.has(name) || (id && exports.coins.has(id)) || (alias && exports.coins.has(alias)));
}
function createTokenMapUsingConfigDetails(tokenConfigMap) {
    const BaseCoins = new Map();
    const nftAndOtherTokens = new Set([
        'erc721:bsctoken',
        'terc721:bsctoken',
        'erc1155:bsctoken',
        'terc1155:bsctoken',
        'erc721:witch',
        'erc721:token',
        'erc1155:token',
        'nonstandard:token',
        'terc721:token',
        'terc1155:token',
        'tnonstandard:token',
        'terc721:bitgoerc721',
        'terc1155:bitgoerc1155',
        'terc721:hoodidj',
        'erc721:polygontoken',
        'erc1155:polygontoken',
        'terc721:polygontoken',
        'terc1155:polygontoken',
        'erc721:soneiumtoken',
        'erc1155:soneiumtoken',
        'terc721:soneiumtoken',
        'terc1155:soneiumtoken',
    ]);
    // Add all the coins from statics coin map first
    exports.coins.forEach((coin, coinName) => {
        BaseCoins.set(coinName, coin);
    });
    // add the tokens not present in the static coin map
    for (const tokenConfigs of Object.values(tokenConfigMap)) {
        const tokenConfig = tokenConfigs[0];
        if (!isCoinPresentInCoinMap({ ...tokenConfig }) &&
            tokenConfig.isToken &&
            !nftAndOtherTokens.has(tokenConfig.name)) {
            const token = createToken(tokenConfig);
            if (token) {
                BaseCoins.set(token.name, token);
            }
        }
    }
    return map_1.CoinMap.fromCoins(Array.from(BaseCoins.values()));
}
function createTokenMapUsingTrimmedConfigDetails(reducedTokenConfigMap) {
    const amsTokenConfigMap = {};
    const networkNameMap = new Map(Object.values(networks_1.Networks).flatMap((networkType) => Object.values(networkType).map((network) => [network.name, network])));
    for (const tokenConfigs of Object.values(reducedTokenConfigMap)) {
        const tokenConfig = tokenConfigs[0];
        const network = networkNameMap.get(tokenConfig.network.name);
        if (!isCoinPresentInCoinMap({ ...tokenConfig }) &&
            network &&
            tokenConfig.isToken &&
            networkFeatureMapForTokens_1.networkFeatureMapForTokens[network.family]) {
            const features = new Set([
                ...(networkFeatureMapForTokens_1.networkFeatureMapForTokens[network.family] || []),
                ...(tokenConfig.additionalFeatures || []),
            ]);
            tokenConfig.excludedFeatures?.forEach((feature) => features.delete(feature));
            amsTokenConfigMap[tokenConfig.name] = [{ ...tokenConfig, features: Array.from(features), network }];
        }
    }
    return createTokenMapUsingConfigDetails(amsTokenConfigMap);
}
function createTokenUsingTrimmedConfigDetails(tokenConfig) {
    let fullTokenConfig;
    const networkNameMap = new Map(Object.values(networks_1.Networks).flatMap((networkType) => Object.values(networkType).map((network) => [network.name, network])));
    const network = networkNameMap.get(tokenConfig.network.name);
    if (!isCoinPresentInCoinMap({ ...tokenConfig }) &&
        network &&
        tokenConfig.isToken &&
        networkFeatureMapForTokens_1.networkFeatureMapForTokens[network.family]) {
        const features = new Set([
            ...(networkFeatureMapForTokens_1.networkFeatureMapForTokens[network.family] || []),
            ...(tokenConfig.additionalFeatures || []),
        ]);
        tokenConfig.excludedFeatures?.forEach((feature) => features.delete(feature));
        fullTokenConfig = { ...tokenConfig, features: Array.from(features), network };
        return createToken(fullTokenConfig);
    }
}
