// @bitgo/statics/dist/src/coinFeatures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WETH_FEATURES = exports.MATIC_FEATURES_WITH_FRANKFURT = exports.MATIC_FEATURES = exports.NEAR_TOKEN_FEATURES = exports.NEAR_FEATURES = exports.STX_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.STX_TOKEN_FEATURES = exports.STX_FEATURES = exports.BSC_TOKEN_FEATURES_EXCLUDE_MENA_FZE = exports.BSC_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.BSC_TOKEN_FEATURES = exports.BSC_FEATURES = exports.APT_OFC_TOKEN_FEATURES = exports.SOL_OFC_TOKEN_FEATURES = exports.SOL_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.SOL_TOKEN_FEATURES = exports.TSOL_FEATURES = exports.SOL_FEATURES = exports.POL_FEATURES = exports.POLYGON_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.POLYGON_TOKEN_FEATURES = exports.POLYGON_FEATURES = exports.HBAR_FEATURES = exports.EOS_FEATURES = exports.DOT_FEATURES = exports.ADA_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.ADA_TOKEN_FEATURES = exports.ADA_FEATURES = exports.HTETH_TOKEN_FEATURES = exports.ALGO_FEATURES = exports.CSPR_FEATURES = exports.POLYGON_TOKEN_FEATURES_WITH_FRANKFURT = exports.XRP_FEATURES = exports.XTZ_FEATURES = exports.XLM_FEATURES = exports.RBTC_FEATURES = exports.CELO_TOKEN_FEATURES = exports.CELO_FEATURES = exports.AVAXC_FEATURES = exports.EVM_FEATURES = exports.ETC_FEATURES = exports.ETH_FEATURES_WITH_STAKING_AND_MMI = exports.ETH_FEATURES_WITH_STAKING = exports.ETH_FEATURES_WITH_MMI = exports.ETH_FEATURES = exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE_AND_MENA_FZE = exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_MENA_FZE = exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE = exports.ACCOUNT_COIN_DEFAULT_FEATURES = exports.BITGO_CUSTODY_JURISDICTIONS = void 0;
exports.BERA_BGT_FEATURES = exports.WCT_FEATURES = exports.TIA_FEATURES = exports.ETH_FEATURES_WITH_FRANKFURT_EXCLUDE_SINGAPORE = exports.ZETA_EVM_FEATURES = exports.ZETA_FEATURES = exports.XLM_TOKEN_FEATURES_WITH_FRANKFURT = exports.SOL_TOKEN_FEATURES_WITH_FRANKFURT_GERMANY = exports.SOL_TOKEN_FEATURES_WITH_FRANKFURT = exports.ETH_FEATURES_WITH_FRANKFURT_GERMANY = exports.ETH_FEATURES_WITH_GERMANY = exports.ETH_FEATURES_WITH_FRANKFURT = exports.POLYX_TOKEN_FEATURES = exports.POLYX_FEATURES = exports.SONEIUM_FEATURES = exports.TAO_TOKEN_FEATURES = exports.TAO_FEATURES = exports.ICP_FEATURES = exports.APT_FEATURES = exports.APECHAIN_FEATURES = exports.TEMPO_FEATURES = exports.COREDAO_FEATURES = exports.OAS_FEATURES = exports.BERA_FEATURES = exports.ZKSYNCERA_FEATURES = exports.ZKETH_FEATURES = exports.OPETH_FEATURES = exports.ARBETH_FEATURES = exports.TON_TOKEN_FEATURES = exports.TON_FEATURES = exports.GENERIC_TOKEN_FEATURES = exports.TOKEN_FEATURES_WITH_NY_GERMANY_FRANKFURT = exports.TOKEN_FEATURES_WITH_FRANKFURT = exports.TOKEN_FEATURES_WITH_SWISS = exports.SEI_FEATURES = exports.COREUM_FEATURES = exports.INJECTIVE_FEATURES = exports.ATOM_FEATURES = exports.COSMOS_TOKEN_FEATURES_WITH_STAKING = exports.COSMOS_TOKEN_FEATURES = exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING = exports.COSMOS_SIDECHAIN_FEATURES = exports.TRX_FEATURES = exports.SUI_TOKEN_FEATURES_STAKING = exports.SUI_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.SUI_TOKEN_FEATURES = exports.SUI_FEATURES = exports.RETH_ROCKET_FEATURES = exports.EIGEN_FEATURES = exports.TWETH_FEATURES = void 0;
exports.CANTON_TOKEN_FEATURES = exports.CANTON_FEATURES = exports.IOTA_FEATURES = exports.WEMIX_FEATURES = exports.WFLR_FEATURES = exports.FLR_FEATURES = exports.SGB_FEATURES = exports.XDC_FEATURES = exports.EVM_NON_EIP1559_FEATURES = exports.VET_TOKEN_FEATURES = exports.VET_FEATURES = void 0;
const base_1 = require("./base");
const ada_1 = require("./ada");
//list of all the jurisdictions where BitGo Custody is available
exports.BITGO_CUSTODY_JURISDICTIONS = [
    base_1.CoinFeature.CUSTODY_BITGO_TRUST,
    base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE,
    base_1.CoinFeature.CUSTODY_BITGO_CUSTODY_MENA_FZE,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.CUSTODY_BITGO_EUROPE_APS,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SISTER_TRUST_ONE,
    base_1.CoinFeature.CUSTODY_BITGO_INDIA,
];
exports.ACCOUNT_COIN_DEFAULT_FEATURES = [
    base_1.CoinFeature.ACCOUNT_MODEL,
    base_1.CoinFeature.REQUIRES_BIG_NUMBER,
    base_1.CoinFeature.VALUELESS_TRANSFER,
    base_1.CoinFeature.TRANSACTION_DATA,
    base_1.CoinFeature.CUSTODY,
    base_1.CoinFeature.CUSTODY_BITGO_TRUST,
    base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE,
    base_1.CoinFeature.CUSTODY_BITGO_CUSTODY_MENA_FZE,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_EUROPE_APS,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_INDIA,
];
exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE = exports.ACCOUNT_COIN_DEFAULT_FEATURES.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE);
exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_MENA_FZE = exports.ACCOUNT_COIN_DEFAULT_FEATURES.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE);
exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE_AND_MENA_FZE = exports.ACCOUNT_COIN_DEFAULT_FEATURES.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE && feature !== base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE);
exports.ETH_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.ENTERPRISE_PAYS_FEES,
    base_1.CoinFeature.WALLET_CONNECT_DEFI,
    base_1.CoinFeature.EVM_COIN,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_ONCHAIN,
    base_1.CoinFeature.SHARED_EVM_MESSAGE_SIGNING,
];
exports.ETH_FEATURES_WITH_MMI = [...exports.ETH_FEATURES, base_1.CoinFeature.METAMASK_INSTITUTIONAL];
exports.ETH_FEATURES_WITH_STAKING = [...exports.ETH_FEATURES, base_1.CoinFeature.STAKING];
exports.ETH_FEATURES_WITH_STAKING_AND_MMI = [...exports.ETH_FEATURES_WITH_STAKING, base_1.CoinFeature.METAMASK_INSTITUTIONAL];
exports.ETC_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.EVM_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.EIP1559,
];
exports.AVAXC_FEATURES = [
    ...exports.ETH_FEATURES_WITH_MMI,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
];
exports.CELO_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.EIP1559,
].filter((feature) => feature !== base_1.CoinFeature.CUSTODY &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_GERMANY &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_TRUST &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_CUSTODY_MENA_FZE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_KOREA &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_EUROPE_APS &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_SISTER_TRUST_ONE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_INDIA &&
    feature !== base_1.CoinFeature.DISTRIBUTED_CUSTODY);
exports.CELO_TOKEN_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES].filter((feature) => feature !== base_1.CoinFeature.CUSTODY &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_TRUST &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_CUSTODY_MENA_FZE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_KOREA &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_EUROPE_APS &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT &&
    feature !== base_1.CoinFeature.CUSTODY_BITGO_INDIA);
exports.RBTC_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
];
exports.XLM_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
];
exports.XTZ_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.ENTERPRISE_PAYS_FEES,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK,
].filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_CUSTODY_MENA_FZE);
exports.XRP_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
];
exports.POLYGON_TOKEN_FEATURES_WITH_FRANKFURT = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.CSPR_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.REQUIRES_RESERVE,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.STAKING,
];
exports.ALGO_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.HTETH_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.WALLET_CONNECT_DEFI,
];
exports.ADA_FEATURES = [
    ...ada_1.Ada.DEFAULT_FEATURES,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
];
exports.ADA_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
];
exports.ADA_TOKEN_FEATURES_EXCLUDE_SINGAPORE = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
];
exports.DOT_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.EXPIRING_TRANSACTIONS,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.EOS_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
];
exports.HBAR_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
];
exports.POLYGON_FEATURES = [
    ...exports.ETH_FEATURES_WITH_MMI,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
];
exports.POLYGON_TOKEN_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.BULK_TRANSACTION];
exports.POLYGON_TOKEN_FEATURES_EXCLUDE_SINGAPORE = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.POL_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
];
exports.SOL_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.REQUIRES_RESERVE,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.WALLET_CONNECT_DEFI,
];
exports.TSOL_FEATURES = [...exports.SOL_FEATURES, base_1.CoinFeature.BULK_TRANSACTION, base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE];
exports.SOL_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.SOL_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.SOL_TOKEN_FEATURES.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE);
exports.SOL_OFC_TOKEN_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.TSS, base_1.CoinFeature.TSS_COLD];
exports.APT_OFC_TOKEN_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.TSS, base_1.CoinFeature.TSS_COLD];
exports.BSC_FEATURES = [
    ...exports.ETH_FEATURES_WITH_STAKING_AND_MMI,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.SHARED_EVM_MESSAGE_SIGNING,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
];
exports.BSC_TOKEN_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.BULK_TRANSACTION];
exports.BSC_TOKEN_FEATURES_EXCLUDE_SINGAPORE = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_SINGAPORE,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.BSC_TOKEN_FEATURES_EXCLUDE_MENA_FZE = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES_EXCLUDE_MENA_FZE,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.STX_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
];
exports.STX_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
];
exports.STX_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.STX_TOKEN_FEATURES.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE);
exports.NEAR_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.SUPPORTS_TOKENS,
];
exports.NEAR_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
];
exports.MATIC_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.METAMASK_INSTITUTIONAL,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
];
exports.MATIC_FEATURES_WITH_FRANKFURT = [...exports.MATIC_FEATURES, base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT];
exports.WETH_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.METAMASK_INSTITUTIONAL,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.LIQUID_STAKING,
];
exports.TWETH_FEATURES = [...exports.WETH_FEATURES, base_1.CoinFeature.STAKING];
exports.EIGEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.METAMASK_INSTITUTIONAL,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.LIQUID_STAKING,
];
exports.RETH_ROCKET_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.METAMASK_INSTITUTIONAL,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.LIQUID_STAKING,
];
exports.SUI_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.BULK_STAKING_TRANSACTION,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.CUSTODY_BULK_TRANSACTION,
    base_1.CoinFeature.ALLOWS_NEGATIVE_FEE,
];
exports.SUI_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.SUI_TOKEN_FEATURES_EXCLUDE_SINGAPORE = exports.SUI_TOKEN_FEATURES.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE);
exports.SUI_TOKEN_FEATURES_STAKING = [
    ...exports.SUI_TOKEN_FEATURES,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.BULK_STAKING_TRANSACTION,
];
exports.TRX_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.SHA256_WITH_ECDSA_TSS,
];
exports.COSMOS_SIDECHAIN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.SHA256_WITH_ECDSA_TSS,
    base_1.CoinFeature.COSMOS_LIKE_COINS,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
    base_1.CoinFeature.INCREASED_TX_REQUEST_REBUILD_LIMIT,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.CUSTODY_BULK_TRANSACTION,
];
exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING = [
    ...exports.COSMOS_SIDECHAIN_FEATURES,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.BULK_STAKING_TRANSACTION,
];
exports.COSMOS_TOKEN_FEATURES = [...exports.COSMOS_SIDECHAIN_FEATURES];
exports.COSMOS_TOKEN_FEATURES_WITH_STAKING = [...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING];
exports.ATOM_FEATURES = [...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING, base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT];
exports.INJECTIVE_FEATURES = [
    ...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
];
exports.COREUM_FEATURES = [...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING, base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT];
exports.SEI_FEATURES = [...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING, base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT];
exports.TOKEN_FEATURES_WITH_SWISS = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND];
exports.TOKEN_FEATURES_WITH_FRANKFURT = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT];
exports.TOKEN_FEATURES_WITH_NY_GERMANY_FRANKFURT = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_NEW_YORK,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
];
exports.GENERIC_TOKEN_FEATURES = [
    base_1.CoinFeature.ACCOUNT_MODEL,
    base_1.CoinFeature.REQUIRES_BIG_NUMBER,
    base_1.CoinFeature.VALUELESS_TRANSFER,
    base_1.CoinFeature.TRANSACTION_DATA,
    base_1.CoinFeature.GENERIC_TOKEN,
];
exports.TON_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BITGO_KOREA,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
    base_1.CoinFeature.SUPPORTS_TOKENS,
];
exports.TON_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
];
exports.ARBETH_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.ETH_ROLLUP_CHAIN,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.WALLET_CONNECT_DEFI,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.OPETH_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.ETH_ROLLUP_CHAIN,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.WALLET_CONNECT_DEFI,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.ZKETH_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.ETH_ROLLUP_CHAIN,
    base_1.CoinFeature.EIP1559,
];
exports.ZKSYNCERA_FEATURES = [
    ...exports.EVM_FEATURES,
    base_1.CoinFeature.SHARED_EVM_SIGNING,
    base_1.CoinFeature.SHARED_EVM_SDK,
    base_1.CoinFeature.EVM_COMPATIBLE_IMS,
    base_1.CoinFeature.EVM_COMPATIBLE_UI,
    base_1.CoinFeature.EVM_COMPATIBLE_WP,
    base_1.CoinFeature.SUPPORTS_ERC20,
    base_1.CoinFeature.EVM_NON_BITGO_RECOVERY,
    base_1.CoinFeature.EVM_UNSIGNED_SWEEP_RECOVERY,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.ETH_ROLLUP_CHAIN,
];
exports.BERA_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
    base_1.CoinFeature.CUSTODY_BULK_TRANSACTION,
];
exports.OAS_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.EIP1559,
];
exports.COREDAO_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
];
exports.TEMPO_FEATURES = [...exports.EVM_FEATURES];
exports.APECHAIN_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.EIP1559,
];
exports.APT_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.BULK_STAKING_TRANSACTION,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.ENTERPRISE_PAYS_FEES,
    base_1.CoinFeature.TSS_ENTERPRISE_PAYS_FEES,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.ICP_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.SHA256_WITH_ECDSA_TSS,
    base_1.CoinFeature.REBUILD_ON_CUSTODY_SIGNING,
];
exports.TAO_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.TAO_TOKEN_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.TSS, base_1.CoinFeature.TSS_COLD];
exports.SONEIUM_FEATURES = [
    ...exports.ETH_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.MPCV2,
    base_1.CoinFeature.EVM_WALLET,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
    base_1.CoinFeature.EIP1559,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG_SUPPORT_GATED,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
];
exports.POLYX_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.STAKING,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.STUCK_TRANSACTION_MANAGEMENT_TSS,
];
exports.POLYX_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TOKEN_STANDARD_USES_ALTERNATIVE_ADDRESS_IDENTIFIER,
];
exports.ETH_FEATURES_WITH_FRANKFURT = [...exports.ETH_FEATURES, base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT];
exports.ETH_FEATURES_WITH_GERMANY = [...exports.ETH_FEATURES, base_1.CoinFeature.CUSTODY_BITGO_GERMANY];
exports.ETH_FEATURES_WITH_FRANKFURT_GERMANY = [...exports.ETH_FEATURES_WITH_FRANKFURT, base_1.CoinFeature.CUSTODY_BITGO_GERMANY];
exports.SOL_TOKEN_FEATURES_WITH_FRANKFURT = [
    ...exports.SOL_TOKEN_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.BULK_TRANSACTION,
];
exports.SOL_TOKEN_FEATURES_WITH_FRANKFURT_GERMANY = [
    ...exports.SOL_TOKEN_FEATURES_WITH_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
];
exports.XLM_TOKEN_FEATURES_WITH_FRANKFURT = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
];
exports.ZETA_FEATURES = [...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING, base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE];
exports.ZETA_EVM_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE];
exports.ETH_FEATURES_WITH_FRANKFURT_EXCLUDE_SINGAPORE = exports.ETH_FEATURES_WITH_FRANKFURT.filter((feature) => feature !== base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE);
exports.TIA_FEATURES = [
    ...exports.COSMOS_SIDECHAIN_FEATURES_WITH_STAKING,
    base_1.CoinFeature.CUSTODY_BITGO_SWITZERLAND,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
];
exports.WCT_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.STAKING, base_1.CoinFeature.WALLET_CONNECT_DEFI];
exports.BERA_BGT_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.CUSTODY_BITGO_GERMANY,
];
exports.VET_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.SUPPORTS_TOKENS,
    base_1.CoinFeature.ENTERPRISE_PAYS_FEES,
    base_1.CoinFeature.TSS_ENTERPRISE_PAYS_FEES,
    base_1.CoinFeature.FEES_PAID_WITH_TOKEN,
    base_1.CoinFeature.MPCV2,
];
exports.VET_TOKEN_FEATURES = exports.VET_FEATURES.filter((feature) => feature !== base_1.CoinFeature.SUPPORTS_TOKENS);
exports.EVM_NON_EIP1559_FEATURES = [...exports.EVM_FEATURES.filter((feature) => feature !== base_1.CoinFeature.EIP1559)];
exports.XDC_FEATURES = [...exports.EVM_NON_EIP1559_FEATURES, base_1.CoinFeature.ERC20_BULK_TRANSACTION];
exports.SGB_FEATURES = [...exports.EVM_FEATURES, base_1.CoinFeature.ERC20_BULK_TRANSACTION];
exports.FLR_FEATURES = [
    ...exports.EVM_FEATURES,
    base_1.CoinFeature.MULTISIG,
    base_1.CoinFeature.MULTISIG_COLD,
    base_1.CoinFeature.MULTISIG_SUPPORT_GATED,
    base_1.CoinFeature.USES_NON_PACKED_ENCODING_FOR_TXDATA,
    base_1.CoinFeature.ERC20_BULK_TRANSACTION,
];
exports.WFLR_FEATURES = [...exports.ACCOUNT_COIN_DEFAULT_FEATURES, base_1.CoinFeature.STABLECOIN, base_1.CoinFeature.STAKING];
exports.WEMIX_FEATURES = [...exports.EVM_FEATURES, base_1.CoinFeature.ERC20_BULK_TRANSACTION];
exports.IOTA_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.BULK_TRANSACTION,
    base_1.CoinFeature.CUSTODY_BULK_TRANSACTION,
];
// TODO: https://bitgoinc.atlassian.net/browse/COIN-5870
exports.CANTON_FEATURES = [
    base_1.CoinFeature.TRANSACTION_DATA,
    base_1.CoinFeature.REQUIRES_BIG_NUMBER,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.CUSTODY,
    base_1.CoinFeature.CUSTODY_BITGO_TRUST,
    base_1.CoinFeature.CUSTODY_BITGO_MENA_FZE,
    base_1.CoinFeature.CUSTODY_BITGO_CUSTODY_MENA_FZE,
    base_1.CoinFeature.CUSTODY_BITGO_SINGAPORE,
    base_1.CoinFeature.CUSTODY_BITGO_EUROPE_APS,
    base_1.CoinFeature.CUSTODY_BITGO_FRANKFURT,
    base_1.CoinFeature.SUPPORTS_ONE_STEP_DEPOSIT,
    base_1.CoinFeature.REQUIRES_WALLET_INITIALIZATION_TRANSACTION,
    base_1.CoinFeature.REQUIRES_DEPOSIT_ACCEPTANCE_TRANSACTION,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
    base_1.CoinFeature.SUPPORTS_TOKENS,
];
exports.CANTON_TOKEN_FEATURES = [
    ...exports.ACCOUNT_COIN_DEFAULT_FEATURES,
    base_1.CoinFeature.TSS,
    base_1.CoinFeature.TSS_COLD,
    base_1.CoinFeature.SUPPORTS_ONE_STEP_DEPOSIT,
    base_1.CoinFeature.REQUIRES_WALLET_INITIALIZATION_TRANSACTION,
    base_1.CoinFeature.REQUIRES_DEPOSIT_ACCEPTANCE_TRANSACTION,
    base_1.CoinFeature.ALPHANUMERIC_MEMO_ID,
];
