// @bitgo/statics/dist/src/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainIdNotFoundError = exports.ConflictingCoinFeaturesError = exports.InvalidDomainError = exports.InvalidContractAddressError = exports.InvalidIdError = exports.MissingRequiredCoinFeatureError = exports.DisallowedCoinFeatureError = exports.DuplicateCoinIdDefinitionError = exports.DuplicateCoinDefinitionError = exports.CoinNotDefinedError = void 0;
class BitGoStaticsError extends Error {
    constructor(message) {
        super(message);
        this.stack = new Error(message).stack;
        Object.setPrototypeOf(this, BitGoStaticsError.prototype);
    }
}
class CoinNotDefinedError extends BitGoStaticsError {
    constructor(coinName) {
        super(`coin '${coinName}' is not defined`);
        Object.setPrototypeOf(this, CoinNotDefinedError.prototype);
    }
}
exports.CoinNotDefinedError = CoinNotDefinedError;
class DuplicateCoinDefinitionError extends BitGoStaticsError {
    constructor(coinName) {
        super(`coin '${coinName}' is already defined`);
        Object.setPrototypeOf(this, DuplicateCoinDefinitionError.prototype);
    }
}
exports.DuplicateCoinDefinitionError = DuplicateCoinDefinitionError;
class DuplicateCoinIdDefinitionError extends BitGoStaticsError {
    constructor(id) {
        super(`coin with id '${id}' is already defined`);
        Object.setPrototypeOf(this, DuplicateCoinIdDefinitionError.prototype);
    }
}
exports.DuplicateCoinIdDefinitionError = DuplicateCoinIdDefinitionError;
class DisallowedCoinFeatureError extends BitGoStaticsError {
    constructor(coinName, feature) {
        super(`coin feature '${feature}' is disallowed for coin ${coinName}.`);
        Object.setPrototypeOf(this, DisallowedCoinFeatureError.prototype);
    }
}
exports.DisallowedCoinFeatureError = DisallowedCoinFeatureError;
class MissingRequiredCoinFeatureError extends BitGoStaticsError {
    constructor(coinName, missingFeatures) {
        super(`Required coin feature(s) '${missingFeatures.join(', ')}' were not found for coin ${coinName}.`);
        Object.setPrototypeOf(this, MissingRequiredCoinFeatureError.prototype);
    }
}
exports.MissingRequiredCoinFeatureError = MissingRequiredCoinFeatureError;
class InvalidIdError extends BitGoStaticsError {
    constructor(coinName, id) {
        super(`invalid uuid '${id}' for coin '${coinName}'`);
        Object.setPrototypeOf(this, InvalidIdError.prototype);
    }
}
exports.InvalidIdError = InvalidIdError;
class InvalidContractAddressError extends BitGoStaticsError {
    constructor(coinName, contractAddress) {
        super(`invalid contract address '${contractAddress}' for coin '${coinName}'`);
        Object.setPrototypeOf(this, InvalidContractAddressError.prototype);
    }
}
exports.InvalidContractAddressError = InvalidContractAddressError;
class InvalidDomainError extends BitGoStaticsError {
    constructor(coinName, domain) {
        super(`invalid domain '${domain}' for coin '${coinName}'`);
        Object.setPrototypeOf(this, InvalidDomainError.prototype);
    }
}
exports.InvalidDomainError = InvalidDomainError;
class ConflictingCoinFeaturesError extends BitGoStaticsError {
    constructor(coinName, conflictingFeatures) {
        super(`coin feature(s) for coin '${coinName}' cannot be both required and disallowed: ${conflictingFeatures.join(', ')}`);
        Object.setPrototypeOf(this, ConflictingCoinFeaturesError.prototype);
    }
}
exports.ConflictingCoinFeaturesError = ConflictingCoinFeaturesError;
class ChainIdNotFoundError extends BitGoStaticsError {
    constructor(chainId) {
        super(`chain '${chainId}' not found`);
        Object.setPrototypeOf(this, CoinNotDefinedError.prototype);
    }
}
exports.ChainIdNotFoundError = ChainIdNotFoundError;
