// @bitgo/statics/dist/src/networkFeatureMapForTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.networkFeatureMapForTokens = void 0;
const account_1 = require("./account");
const base_1 = require("./base");
const coinFeatures_1 = require("./coinFeatures");
exports.networkFeatureMapForTokens = {
    algo: account_1.AccountCoin.DEFAULT_FEATURES,
    apt: coinFeatures_1.APT_FEATURES,
    arbeth: [...account_1.AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559],
    avaxc: account_1.AccountCoin.DEFAULT_FEATURES,
    bera: [...account_1.AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559],
    bsc: coinFeatures_1.BSC_TOKEN_FEATURES,
    celo: account_1.AccountCoin.DEFAULT_FEATURES,
    eth: account_1.AccountCoin.DEFAULT_FEATURES,
    eos: account_1.AccountCoin.DEFAULT_FEATURES,
    hbar: account_1.AccountCoin.DEFAULT_FEATURES,
    opeth: [...account_1.AccountCoin.DEFAULT_FEATURES, base_1.CoinFeature.EIP1559],
    polygon: coinFeatures_1.POLYGON_TOKEN_FEATURES,
    sol: coinFeatures_1.SOL_TOKEN_FEATURES,
    stx: coinFeatures_1.STX_TOKEN_FEATURES,
    sui: coinFeatures_1.SUI_TOKEN_FEATURES,
    tao: coinFeatures_1.TAO_TOKEN_FEATURES,
    polyx: coinFeatures_1.POLYX_TOKEN_FEATURES,
    trx: account_1.AccountCoin.DEFAULT_FEATURES,
    xlm: account_1.AccountCoin.DEFAULT_FEATURES,
    xrp: account_1.AccountCoin.DEFAULT_FEATURES,
    ofc: [],
    ada: coinFeatures_1.ADA_TOKEN_FEATURES,
};
