// @bitgo/statics/dist/src/networks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Networks = exports.FlareTestnet = exports.Flare = exports.FlarePTestnet = exports.FlareP = exports.BaseNetwork = exports.NetworkType = void 0;
const base_1 = require("./base");
var NetworkType;
(function (NetworkType) {
    NetworkType["MAINNET"] = "mainnet";
    NetworkType["TESTNET"] = "testnet";
})(NetworkType || (exports.NetworkType = NetworkType = {}));
class BaseNetwork {
}
exports.BaseNetwork = BaseNetwork;
class Mainnet extends BaseNetwork {
    constructor() {
        super(...arguments);
        this.type = NetworkType.MAINNET;
    }
}
class Testnet extends BaseNetwork {
    constructor() {
        super(...arguments);
        this.type = NetworkType.TESTNET;
    }
}
class Algorand extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Algorand';
        this.family = base_1.CoinFamily.ALGO;
        this.explorerUrl = 'https://allo.info/tx/';
    }
}
class AlgorandTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'AlgorandTestnet';
        this.family = base_1.CoinFamily.ALGO;
        this.explorerUrl = 'https://explorer.bitquery.io/algorand_testnet/tx/';
    }
}
class Ada extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'AdaCardano';
        this.family = base_1.CoinFamily.ADA;
        this.utxolibName = 'cardano';
        this.poolDeposit = 500000000;
        this.stakeKeyDeposit = 2000000;
        this.explorerUrl = 'https://cardanoscan.io/transaction/';
        this.accountExplorerUrl = 'https://cardanoscan.io/address/';
        this.coinsPerUtxoWord = 34482;
        this.maxTransactionSize = 8000;
        this.maxValueSize = 4000;
    }
}
class AdaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'AdaCardanoTestnet';
        this.family = base_1.CoinFamily.ADA;
        this.utxolibName = 'cardanoTestnet';
        this.explorerUrl = 'https://preprod.cardanoscan.io/transaction/';
        this.accountExplorerUrl = 'https://preprod.cardanoscan.io/address/';
        this.coinsPerUtxoWord = 34482;
        this.maxTransactionSize = 8000;
        this.maxValueSize = 4000;
        this.poolDeposit = 500000000;
        this.stakeKeyDeposit = 2000000;
    }
}
class Apt extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Apt';
        this.family = base_1.CoinFamily.APT;
        this.explorerUrl = 'https://explorer.aptoslabs.com/txn/?network=mainnet';
        this.accountExplorerUrl = 'https://explorer.aptoslabs.com/account/?network=mainnet';
        this.blockExplorerUrl = 'https://explorer.aptoslabs.com/block/?network=mainnet';
    }
}
class AptTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Apt';
        this.family = base_1.CoinFamily.APT;
        this.explorerUrl = 'https://explorer.aptoslabs.com/txn/?network=testnet';
        this.accountExplorerUrl = 'https://explorer.aptoslabs.com/account/?network=testnet';
        this.blockExplorerUrl = 'https://explorer.aptoslabs.com/block/?network=testnet';
    }
}
class Icp extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Internet Computer';
        this.family = base_1.CoinFamily.ICP;
        this.explorerUrl = 'https://dashboard.internetcomputer.org/transaction/';
        this.accountExplorerUrl = 'https://dashboard.internetcomputer.org/account/';
    }
}
class IcpTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Internet Computer';
        this.family = base_1.CoinFamily.ICP;
        this.explorerUrl = 'https://dashboard.internetcomputer.org/transaction/'; // ICP testnet explorer is not available, so using mainnet explorer
        this.accountExplorerUrl = 'https://dashboard.internetcomputer.org/account/';
    }
}
class Arbitrum extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Arbitrum';
        this.family = base_1.CoinFamily.ARBETH;
        this.explorerUrl = 'https://arbiscan.io/tx/';
        this.accountExplorerUrl = 'https://arbiscan.io/address/';
        this.chainId = 42161;
        this.nativeCoinOperationHashPrefix = '42161';
        this.tokenOperationHashPrefix = '42161-ERC20';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class ArbitrumTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ArbitrumTestnet';
        this.family = base_1.CoinFamily.ARBETH;
        this.explorerUrl = 'https://sepolia.arbiscan.io/tx/';
        this.accountExplorerUrl = 'https://sepolia.arbiscan.io/address/';
        this.chainId = 421614;
        this.nativeCoinOperationHashPrefix = '421614';
        this.tokenOperationHashPrefix = '421614-ERC20';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class AvalancheC extends Mainnet {
    constructor() {
        super(...arguments);
        // https://docs.avax.network/build/tutorials/smart-contracts/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask#avalanche-mainnet-settings
        this.name = 'AvalancheC';
        this.family = base_1.CoinFamily.AVAXC;
        this.explorerUrl = 'https://subnets.avax.network/c-chain/tx/';
        this.accountExplorerUrl = 'https://subnets.avax.network/c-chain/address/';
        this.chainId = 43114;
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class AvalancheCTestnet extends Testnet {
    constructor() {
        super(...arguments);
        // https://docs.avax.network/build/tutorials/smart-contracts/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask#avalanche-mainnet-settings
        this.name = 'AvalancheCTestnet';
        this.family = base_1.CoinFamily.AVAXC;
        this.explorerUrl = 'https://subnets-test.avax.network/c-chain/tx/';
        this.accountExplorerUrl = 'https://subnets-test.avax.network/c-chain/address/';
        this.chainId = 43113;
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class AvalancheP extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'AvalancheP';
        this.family = base_1.CoinFamily.AVAXP;
        this.explorerUrl = 'https://subnets.avax.network/p-chain/tx/';
        this.accountExplorerUrl = 'https://subnets.avax.network/p-chain/address/';
        this.blockchainID = '11111111111111111111111111111111LpoYY';
        this.cChainBlockchainID = '2q9e4r6Mu3U68nU1fYjgbR6JvwrRx36CohpAX5UQxse55x1Q5';
        this.avaxAssetID = 'FvwEAhmxKfeiG8SnEvq42hc6whRyY3EFYAvebMqDNDGCgxN5Z';
        this.networkID = 1;
        this.hrp = 'avax';
        this.alias = 'P';
        this.vm = 'platformvm';
        this.txFee = '1000000'; // 1 MILLIAVAX
        this.maxImportFee = '10000000'; // 0.01 AVAX
        this.createSubnetTx = '1000000000'; // 1 AVAX
        this.createChainTx = '1000000000'; // 1 AVAX
        this.creationTxFee = '10000000'; // 1 CENTIAVAX
        this.minConsumption = '0.1';
        this.maxConsumption = '0.12';
        this.maxSupply = '720000000000000000'; // 720 mil tokens
        this.minStake = '2000000000000'; // 2000 AVAX
        this.minStakeDuration = '1209600'; // 2 weeks
        this.maxStakeDuration = '31536000'; // 1 year
        this.minDelegationStake = '25000000000'; // 25 AVAX
        this.minDelegationFee = '2';
    }
}
class AvalanchePTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'AvalanchePTestnet';
        this.family = base_1.CoinFamily.AVAXP;
        this.explorerUrl = 'https://subnets-test.avax.network/p-chain/tx/';
        this.accountExplorerUrl = 'https://subnets-test.avax.network/p-chain/address/';
        this.blockchainID = '11111111111111111111111111111111LpoYY';
        this.cChainBlockchainID = 'yH8D7ThNJkxmtkuv2jgBa4P1Rn3Qpr4pPr7QYNfcdoS6k6HWp';
        this.avaxAssetID = 'U8iRqJoiJm8xZHAacmvYyZVwqQx6uDNtQeP3CQ6fcgQk3JqnK';
        this.networkID = 5;
        this.alias = 'P';
        this.hrp = 'fuji';
        this.vm = 'platformvm';
        this.txFee = '1000000'; // 1 MILLIAVAX
        this.maxImportFee = '10000000'; // 0.01 AVAX
        this.createSubnetTx = '1000000000'; // 1 AVAX
        this.createChainTx = '1000000000'; // 1 AVAX
        this.creationTxFee = '10000000'; // 1 CENTIAVAX
        this.minConsumption = '0.1';
        this.maxConsumption = '0.12';
        this.maxSupply = '720000000000000000'; // 720 mil tokens
        this.minStake = '1000000000'; // 1 AVAX
        this.minStakeDuration = '86400'; // 1 day
        this.maxStakeDuration = '31536000'; // 1 year
        this.minDelegationStake = '1000000000'; // 1 AVAX
        this.minDelegationFee = '2';
    }
}
class BinanceSmartChain extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'BinanceSmartChain';
        this.family = base_1.CoinFamily.BSC;
        this.explorerUrl = 'https://bscscan.com/tx/';
        this.accountExplorerUrl = 'https://bscscan.com/address/';
        this.chainId = 56;
        this.nativeCoinOperationHashPrefix = '56';
        this.tokenOperationHashPrefix = '56-ERC20';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class BinanceSmartChainTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BinanceSmartChainTestnet';
        this.family = base_1.CoinFamily.BSC;
        this.explorerUrl = 'https://testnet.bscscan.com/tx/';
        this.accountExplorerUrl = 'https://testnet.bscscan.com/address/';
        this.chainId = 97;
        this.nativeCoinOperationHashPrefix = '97';
        this.tokenOperationHashPrefix = '97-ERC20';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class OpBNB extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'OpBNB';
        this.family = base_1.CoinFamily.OPBNB;
        this.explorerUrl = 'https://opbnbscan.com/tx/';
        this.accountExplorerUrl = 'https://opbnbscan.com/address/';
        this.chainId = 204;
        this.nativeCoinOperationHashPrefix = '204';
        this.tokenOperationHashPrefix = '204-ERC20';
    }
}
class OpBNBTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'OpBNBTestnet';
        this.family = base_1.CoinFamily.OPBNB;
        this.explorerUrl = 'https://testnet.opbnbscan.com/tx/';
        this.accountExplorerUrl = 'https://testnet.opbnbscan.com/address/';
        this.chainId = 5611;
        this.nativeCoinOperationHashPrefix = '5611';
        this.tokenOperationHashPrefix = '5611-ERC20';
    }
}
class Fantom extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Fantom';
        this.family = base_1.CoinFamily.FANTOM;
        this.explorerUrl = 'https://explorer.fantom.network/transactions/';
        this.accountExplorerUrl = 'https://explorer.fantom.network/address/';
        this.chainId = 250;
        this.nativeCoinOperationHashPrefix = '250';
        this.tokenOperationHashPrefix = '250-ERC20';
    }
}
class FantomTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'FantomTestnet';
        this.family = base_1.CoinFamily.FANTOM;
        this.explorerUrl = 'https://explorer.fantom.network/transactions/';
        this.accountExplorerUrl = 'https://explorer.fantom.network/address/';
        this.chainId = 4002;
        this.nativeCoinOperationHashPrefix = '4002';
        this.tokenOperationHashPrefix = '4002-ERC20';
    }
}
class LightningBitcoin extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'LightningBitcoin';
        this.family = base_1.CoinFamily.LNBTC;
        this.utxolibName = 'bitcoin';
        this.explorerUrl = 'https://mempool.space/lightning';
        this.lightningServicePubKey = '0338508686f978ceffd7ce05404041b1a5b4f75a39bc92a6d355240ccc081f763e';
        this.middlewarePubKey = '020d52f6458b17b565bc03516568d1a65c8cbd131ce03f2207567e4cd3322e82f3';
        this.tatPubKey = '03953da20068c096858d3cca60f8c5b0e5a2f5d0383ecf9b6ab392d71e4f3e72df';
    }
}
class LightningBitcoinTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'LightningBitcoinTestnet';
        this.family = base_1.CoinFamily.LNBTC;
        this.utxolibName = 'testnet';
        this.explorerUrl = 'https://mempool.space/testnet/lightning';
        this.lightningServicePubKey = '024055021db1e7f019ebb783ab0b0810c21a819207d4cb1ec4a6e2150ac07f1482';
        this.middlewarePubKey = '027cb3bc6b49fc385d282b42a7be232a94ffcbaffc7818b603b17722582bbf539b';
        this.tatPubKey = '02e747c99c371eac9c14fb19913bec8a0e3e46e35ab1a45878e5b9afbb69899c1e';
    }
}
class Bitcoin extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Bitcoin';
        this.family = base_1.CoinFamily.BTC;
        this.utxolibName = 'bitcoin';
        this.explorerUrl = 'https://mempool.space/tx/';
    }
}
class BitcoinTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinTestnet';
        this.family = base_1.CoinFamily.BTC;
        this.utxolibName = 'testnet';
        this.explorerUrl = 'https://mempool.space/testnet/tx/';
        // Add our pubkey for our paygo wallet
        this.paygoAddressAttestationPubkey = 'xpub6BKRgmCPX5oQiJgJ6Vq6BF8tDvZhwQki5dVVQohckK2ZJXtxj8K6M9pavLwt9piW33hZz17SWmG8QWsjJ1tHdde2Fs5UA3DFbApCtbdaGKn';
    }
}
class BitcoinPublicSignet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinPublicSignet';
        this.family = base_1.CoinFamily.BTC;
        this.utxolibName = 'bitcoinPublicSignet';
        this.explorerUrl = 'https://mempool.space/signet/tx/';
    }
}
class BitcoinTestnet4 extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinTestnet4';
        this.family = base_1.CoinFamily.BTC;
        this.utxolibName = 'bitcoinTestnet4';
        this.explorerUrl = 'https://mempool.space/testnet4/tx/';
    }
}
class BitcoinBitGoSignet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinBitGoSignet';
        this.family = base_1.CoinFamily.BTC;
        this.utxolibName = 'bitcoinBitGoSignet';
        this.explorerUrl = 'https://mempool.space/notanetwork/tx/';
    }
}
class BitcoinCash extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinCash';
        this.family = base_1.CoinFamily.BCH;
        this.utxolibName = 'bitcoincash';
        this.explorerUrl = 'https://www.blockchain.com/bch/tx/';
    }
}
class BitcoinCashTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinCashTestnet';
        this.family = base_1.CoinFamily.BCH;
        this.utxolibName = 'bitcoincashTestnet';
        this.explorerUrl = 'https://tbch.loping.net/tx/';
    }
}
class BitcoinSV extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinSV';
        this.family = base_1.CoinFamily.BSV;
        this.utxolibName = 'bitcoinsv';
        this.explorerUrl = 'https://blockchair.com/bitcoin-sv/transaction/';
    }
}
class BitcoinSVTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinSVTestnet';
        this.family = base_1.CoinFamily.BSV;
        this.utxolibName = 'bitcoinsvTestnet';
        this.explorerUrl = undefined;
    }
}
class BitcoinGold extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinGold';
        this.family = base_1.CoinFamily.BTG;
        this.utxolibName = 'bitcoingold';
        this.explorerUrl = 'https://btgexplorer.com/tx/';
    }
}
class BitcoinGoldTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BitcoinGoldTestnet';
        this.family = base_1.CoinFamily.BTG;
        this.utxolibName = 'bitcoingoldTestnet';
        this.explorerUrl = undefined;
    }
}
class Dash extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Dash';
        this.family = base_1.CoinFamily.DASH;
        this.utxolibName = 'dash';
        this.explorerUrl = 'https://insight.dashevo.org/insight/tx/';
    }
}
class DashTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'DashTestnet';
        this.family = base_1.CoinFamily.DASH;
        this.utxolibName = 'dashTest';
        this.explorerUrl = 'https://testnet-insight.dashevo.org/insight/tx/';
    }
}
class Dogecoin extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Dogecoin';
        this.family = base_1.CoinFamily.DOGE;
        this.utxolibName = 'dogecoin';
        this.explorerUrl = 'https://blockchair.com/dogecoin/transaction/';
    }
}
class DogecoinTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'DogecoinTestnet';
        this.family = base_1.CoinFamily.DOGE;
        this.utxolibName = 'dogecoinTest';
        this.explorerUrl = 'https://blockexplorer.one/dogecoin/testnet/tx/';
    }
}
class ECash extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'ECash';
        this.family = base_1.CoinFamily.BCHA;
        this.utxolibName = 'ecash';
        this.explorerUrl = 'https://blockchair.com/ecash';
    }
}
class ECashTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ECashTestnet';
        this.family = base_1.CoinFamily.BCHA;
        this.utxolibName = 'ecashTest';
        this.explorerUrl = undefined;
    }
}
class PolkadotAssetHub extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'PolkadotAssetHub';
        this.family = base_1.CoinFamily.DOT;
        this.explorerUrl = 'https://assethub-polkadot.subscan.io/extrinsic/';
        this.specName = 'statemint';
        this.genesisHash = '0x68d56f15f85d3136970ec16946040bc1752654e906147f7e43e9d539d7c3de2f';
        this.specVersion = 1007001;
        this.chainName = 'Polkadot Asset Hub';
        this.txVersion = 15;
    }
}
class WestendAssetHub extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'WestendAssetHub';
        this.family = base_1.CoinFamily.DOT;
        this.explorerUrl = 'https://assethub-westend.subscan.io/extrinsic/';
        this.specName = 'westmint';
        this.genesisHash = '0x67f9723393ef76214df0118c34bbbd3dbebc8ed46a10973a8c969d48fe7598c9';
        this.specVersion = 1020000;
        this.chainName = 'Westend Asset Hub';
        this.txVersion = 16;
    }
}
class Celo extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Celo';
        this.family = base_1.CoinFamily.CELO;
        this.explorerUrl = 'https://explorer.celo.org/tx/';
        this.accountExplorerUrl = 'https://explorer.celo.org/address/';
        this.chainId = 42220;
        this.nativeCoinOperationHashPrefix = 'CELO';
        this.tokenOperationHashPrefix = 'CELO-ERC20';
    }
}
class CeloTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'CeloTestnet';
        this.family = base_1.CoinFamily.CELO;
        this.explorerUrl = 'https://sepolia.celoscan.io/tx/';
        this.accountExplorerUrl = 'https://sepolia.celoscan.io/address/';
        this.chainId = 11142220;
        this.nativeCoinOperationHashPrefix = 'CELO';
        this.tokenOperationHashPrefix = 'CELO-ERC20';
    }
}
class Chiliz extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Chiliz';
        this.family = base_1.CoinFamily.CHILIZ;
        this.explorerUrl = 'https://chiliscan.com/tx/';
        this.accountExplorerUrl = 'https://chiliscan.com/address/';
        this.chainId = 88888;
        this.nativeCoinOperationHashPrefix = '88888';
        this.tokenOperationHashPrefix = '88888-ERC20';
    }
}
class ChilizTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ChilizTestnet';
        this.family = base_1.CoinFamily.CHILIZ;
        this.explorerUrl = 'https://testnet.chiliscan.com/tx/';
        this.accountExplorerUrl = 'https://testnet.chiliscan.com/address/';
        this.chainId = 88882;
        this.nativeCoinOperationHashPrefix = '88882';
        this.tokenOperationHashPrefix = '88882-ERC20';
    }
}
class CodexEth extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Codex Ethereum';
        this.family = base_1.CoinFamily.CODEXETH;
        this.explorerUrl = 'https://explorer.codex.xyz/tx/';
        this.accountExplorerUrl = 'https://explorer.codex.xyz/address/';
        this.chainId = 81224;
        this.nativeCoinOperationHashPrefix = '81224';
    }
}
class CodexEthTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Codex Ethereum';
        this.family = base_1.CoinFamily.CODEXETH;
        this.explorerUrl = 'https://explorer.codex-stg.xyz/tx/';
        this.accountExplorerUrl = 'https://explorer.codex-stg.xyz/address/';
        this.chainId = 812242;
        this.nativeCoinOperationHashPrefix = '812242';
    }
}
// TODO update explorerUrl STLX-1657
class Casper extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Casper';
        this.family = base_1.CoinFamily.CSPR;
        this.explorerUrl = 'https://cspr.live/deploy/';
        this.accountExplorerUrl = 'https://cspr.live/account/';
    }
}
class CasperTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'CasperTestnet';
        this.family = base_1.CoinFamily.CSPR;
        this.explorerUrl = 'https://integration.cspr.live/deploy/';
        this.accountExplorerUrl = 'https://integration.cspr.live/account/';
    }
}
class Ethereum extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Ethereum';
        this.family = base_1.CoinFamily.ETH;
        this.explorerUrl = 'https://etherscan.io/tx/';
        this.accountExplorerUrl = 'https://etherscan.io/address/';
        this.blockExplorerUrl = 'https://etherscan.io/block/';
        // from https://github.com/ethereumjs/ethereumjs-common/blob/a978f630858f6843176bb20b277569785914e899/src/chains/index.ts
        this.chainId = 1;
        this.batcherContractAddress = '0xebe27913fcc7510eadf10643a8f86bf5492a9541';
        this.forwarderFactoryAddress = '0xffa397285ce46fb78c588a9e993286aac68c37cd';
        this.forwarderImplementationAddress = '0x059ffafdc6ef594230de44f824e2bd0a51ca5ded';
        this.walletV2ForwarderFactoryAddress = '0x29Ef46035e9fA3D570c598d3266424Ca11413b0C';
        this.walletV2ForwarderImplementationAddress = '0x5397d0869aBA0D55e96D5716d383F6e1d8695ed7';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.nativeCoinOperationHashPrefix = 'ETHER';
        this.tokenOperationHashPrefix = 'ERC20';
        this.walletFactoryAddress = '0x9a0d63911620f7fc15c3c020edbe4d7267ea3e4d';
        this.walletImplementationAddress = '0xe8e847cf573fc8ed75621660a36affd18c543d7e';
        this.walletV2FactoryAddress = '0xa7198f48C58E91f01317E70Cd24C5Cce475c1555';
        this.walletV2ImplementationAddress = '0xe5DcdC13B628c2df813DB1080367E929c1507Ca0';
        this.walletV4FactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletV4ImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Ethereum2 extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Ethereum2';
        this.family = base_1.CoinFamily.ETH2;
        this.explorerUrl = 'https://beaconscan.com/tx';
        this.accountExplorerUrl = 'https://beaconscan.com/address';
    }
}
class EthereumW extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Ethereum PoW';
        this.family = base_1.CoinFamily.ETHW;
        this.explorerUrl = '';
        this.accountExplorerUrl = '';
        this.chainId = 10001;
        this.batcherContractAddress = '';
        this.forwarderFactoryAddress = '';
        this.forwarderImplementationAddress = '';
        this.nativeCoinOperationHashPrefix = 'ETHER';
        this.tokenOperationHashPrefix = 'ERC20';
    }
}
class XtzEvm extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'XTZ EVM';
        this.family = base_1.CoinFamily.XTZEVM;
        this.explorerUrl = 'https://explorer.etherlink.com/tx/';
        this.accountExplorerUrl = 'https://explorer.etherlink.com/address/';
        this.chainId = 42793;
        this.nativeCoinOperationHashPrefix = '42793';
    }
}
class XtzEvmTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet XTZ EVM';
        this.family = base_1.CoinFamily.XTZEVM;
        this.explorerUrl = 'https://shadownet.explorer.etherlink.com/tx/';
        this.accountExplorerUrl = 'https://shadownet.explorer.etherlink.com/address/';
        this.chainId = 128123;
        this.nativeCoinOperationHashPrefix = '128123';
    }
}
class Pyrmont extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Pyrmont';
        this.family = base_1.CoinFamily.ETH2;
        this.explorerUrl = 'https://beaconscan.com/pyrmont/tx';
        this.accountExplorerUrl = 'https://beaconscan.com/pyrmont/address';
    }
}
class Kovan extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Kovan';
        this.family = base_1.CoinFamily.ETH;
        this.explorerUrl = 'https://kovan.etherscan.io/tx/';
        this.accountExplorerUrl = 'https://kovan.etherscan.io/address/';
        // from https://github.com/ethereumjs/ethereumjs-common/blob/a978f630858f6843176bb20b277569785914e899/src/chains/index.ts
        this.chainId = 42;
        this.batcherContractAddress = '0xc0aaf2649e7b0f3950164681eca2b1a8f654a478';
        this.forwarderFactoryAddress = '0xa79a485294d226075ee65410bc94ea454f3e409d';
        this.forwarderImplementationAddress = '0xa946e748f25a5ec6878eb1a9f2e902028174c0b3';
        this.nativeCoinOperationHashPrefix = 'ETHER';
        this.tokenOperationHashPrefix = 'ERC20';
    }
}
class Goerli extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Goerli';
        this.family = base_1.CoinFamily.ETH;
        this.explorerUrl = 'https://goerli.etherscan.io/tx/';
        this.accountExplorerUrl = 'https://goerli.etherscan.io/address/';
        this.blockExplorerUrl = 'https://goerli.etherscan.io/block/';
        // from https://github.com/ethereumjs/ethereumjs-common/blob/a978f630858f6843176bb20b277569785914e899/src/chains/index.ts
        this.chainId = 5;
        this.batcherContractAddress = '0xe8e847cf573fc8ed75621660a36affd18c543d7e';
        this.forwarderFactoryAddress = '0xf5caa5e3e93afbc21bd19ef4f2691a37121f7917';
        this.forwarderImplementationAddress = '0x80d5c91e8cc21df69fc4d64f21dc2d83121c3999';
        this.nativeCoinOperationHashPrefix = 'ETHER';
        this.tokenOperationHashPrefix = 'ERC20';
    }
}
class Holesky extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Holesky';
        this.family = base_1.CoinFamily.ETH;
        this.explorerUrl = 'https://holesky.etherscan.io/tx/';
        this.accountExplorerUrl = 'https://holesky.etherscan.io/address/';
        this.blockExplorerUrl = 'https://holesky.etherscan.io/block/';
        // https://chainlist.org/chain/17000
        this.chainId = 17000;
        this.batcherContractAddress = '0xebe27913fcc7510eadf10643a8f86bf5492a9541';
        this.forwarderFactoryAddress = '0xffa397285ce46fb78c588a9e993286aac68c37cd';
        this.forwarderImplementationAddress = '0x059ffafdc6ef594230de44f824e2bd0a51ca5ded';
        this.nativeCoinOperationHashPrefix = 'ETHER';
        this.tokenOperationHashPrefix = 'ERC20';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class Hoodi extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Hoodi';
        this.family = base_1.CoinFamily.ETH;
        this.explorerUrl = 'https://hoodi.etherscan.io/tx/';
        this.accountExplorerUrl = 'https://hoodi.etherscan.io/address/';
        this.blockExplorerUrl = 'https://hoodi.etherscan.io/block/';
        // https://chainlist.org/chain/560048
        this.chainId = 560048;
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x0e2874d6824fae4f61e446012317a9b86384bd8e';
        this.forwarderImplementationAddress = '0x7441f20a59be97011842404b9aefd8d85fd81aa6';
        this.walletV2ForwarderFactoryAddress = '0x0e2874d6824fae4f61e446012317a9b86384bd8e';
        this.walletV2ForwarderImplementationAddress = '0x7441f20a59be97011842404b9aefd8d85fd81aa6';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0xf514cd80a41bde2e20033b251f1f74633caf3e59';
        this.walletImplementationAddress = '0x6bb86b3b27b092bf8a285080fe7d58acdf841041';
        this.walletV2FactoryAddress = '0xf514cd80a41bde2e20033b251f1f74633caf3e59';
        this.walletV2ImplementationAddress = '0x6bb86b3b27b092bf8a285080fe7d58acdf841041';
        this.walletV4FactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletV4ImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.nativeCoinOperationHashPrefix = 'ETHER';
        this.tokenOperationHashPrefix = 'ERC20';
    }
}
class EthereumClassic extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'EthereumClassic';
        this.family = base_1.CoinFamily.ETC;
        this.explorerUrl = 'https://blockscout.com/etc/mainnet/tx/';
        this.accountExplorerUrl = 'https://blockscout.com/etc/mainnet/address/';
        // from  https://chainid.network/chains/
        this.chainId = 61;
        this.nativeCoinOperationHashPrefix = 'ETC';
        this.tokenOperationHashPrefix = 'ETC-ERC20';
    }
}
class EthereumClassicTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'EthereumClassicTestnet';
        this.family = base_1.CoinFamily.ETC;
        this.explorerUrl = 'https://blockscout.com/etc/mordor/tx/';
        this.accountExplorerUrl = 'https://blockscout.com/etc/mordor/address/';
        // from  https://chainid.network/chains/
        this.chainId = 63;
        this.nativeCoinOperationHashPrefix = 'ETC';
        this.tokenOperationHashPrefix = 'ETC-ERC20';
    }
}
class Eos extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Eos';
        this.family = base_1.CoinFamily.EOS;
        this.explorerUrl = 'https://explorer.eosnetwork.com/transaction/';
    }
}
class EosTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'EosTestnet';
        this.family = base_1.CoinFamily.EOS;
        this.explorerUrl = 'https://kylin.eosx.io/tx/';
    }
}
class Hedera extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Hedera';
        this.family = base_1.CoinFamily.HBAR;
        this.explorerUrl = 'https://hashscan.io/mainnet/transaction/';
    }
}
class HederaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'HederaTestnet';
        this.family = base_1.CoinFamily.HBAR;
        this.explorerUrl = 'https://testnet.dragonglass.me/hedera/search?q=';
    }
}
class Litecoin extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Litecoin';
        this.family = base_1.CoinFamily.LTC;
        this.utxolibName = 'litecoin';
        this.explorerUrl = 'https://litecoinspace.org/tx/';
    }
}
class LitecoinTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'LitecoinTestnet';
        this.family = base_1.CoinFamily.LTC;
        this.utxolibName = 'litecoinTest';
        this.explorerUrl = 'https://blockexplorer.one/litecoin/testnet/tx/';
    }
}
class Ofc extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Ofc';
        this.family = base_1.CoinFamily.OFC;
        this.explorerUrl = undefined;
    }
}
class OfcTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'OfcTestnet';
        this.family = base_1.CoinFamily.OFC;
        this.explorerUrl = undefined;
    }
}
class Rbtc extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Rbtc';
        this.family = base_1.CoinFamily.RBTC;
        this.explorerUrl = 'https://explorer.rsk.co/tx/';
        this.accountExplorerUrl = 'https://explorer.rsk.co/address/';
        this.chainId = 30;
        this.nativeCoinOperationHashPrefix = 'RSK';
        this.tokenOperationHashPrefix = 'RSK-ERC20';
    }
}
class RbtcTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'RbtcTestnet';
        this.family = base_1.CoinFamily.RBTC;
        this.explorerUrl = 'https://explorer.testnet.rsk.co/tx/';
        this.accountExplorerUrl = 'https://explorer.testnet.rsk.co/address/';
        this.chainId = 31;
        this.nativeCoinOperationHashPrefix = 'RSK';
        this.tokenOperationHashPrefix = 'RSK-ERC20';
    }
}
class Stellar extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Stellar';
        this.family = base_1.CoinFamily.XLM;
        this.explorerUrl = 'https://stellar.expert/explorer/public/tx/';
    }
}
class StellarTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'StellarTestnet';
        this.family = base_1.CoinFamily.XLM;
        this.explorerUrl = 'https://stellar.expert/explorer/testnet/tx/';
    }
}
class Sol extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Sol';
        this.family = base_1.CoinFamily.SOL;
        this.explorerUrl = 'https://explorer.solana.com/tx/';
    }
}
class SolTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SolTestnet';
        this.family = base_1.CoinFamily.SOL;
        this.explorerUrl = 'https://explorer.solana.com/tx/?cluster=devnet';
    }
}
class Sui extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Sui';
        this.family = base_1.CoinFamily.SUI;
        this.explorerUrl = 'https://explorer.sui.io/txblock/?network=mainnet';
        this.accountExplorerUrl = 'https://explorer.sui.io/address/?network=mainnet';
    }
}
class SuiTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Sui';
        this.family = base_1.CoinFamily.SUI;
        this.explorerUrl = 'https://explorer.sui.io/txblock/?network=testnet';
        this.accountExplorerUrl = 'https://explorer.sui.io/address/?network=testnet';
    }
}
class Atom extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Cosmos Hub ATOM';
        this.family = base_1.CoinFamily.ATOM;
        this.explorerUrl = 'https://www.mintscan.io/cosmos/tx/';
    }
}
class AtomTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Cosmos Hub ATOM';
        this.family = base_1.CoinFamily.ATOM;
        this.explorerUrl = 'https://explorer.polypore.xyz/provider/tx/';
    }
}
class Osmo extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Osmosis';
        this.family = base_1.CoinFamily.OSMO;
        this.explorerUrl = 'https://www.mintscan.io/osmosis/txs/';
    }
}
class OsmoTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Osmosis';
        this.family = base_1.CoinFamily.OSMO;
        this.explorerUrl = 'https://testnet.osmosis.explorers.guru/transaction/';
    }
}
class Tia extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Celestia';
        this.family = base_1.CoinFamily.TIA;
        this.explorerUrl = 'https://www.mintscan.io/celestia/tx/';
    }
}
class TiaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Celestia';
        this.family = base_1.CoinFamily.TIA;
        this.explorerUrl = 'https://testnet.celestia.explorers.guru/transaction/';
    }
}
class Hash extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Provenance';
        this.family = base_1.CoinFamily.HASH;
        this.explorerUrl = 'https://explorer.provenance.io/tx/';
    }
}
class HashTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Provenance';
        this.family = base_1.CoinFamily.HASH;
        this.explorerUrl = 'https://explorer.test.provenance.io/tx/';
    }
}
class Bld extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Agoric';
        this.family = base_1.CoinFamily.BLD;
        this.explorerUrl = 'https://bigdipper.live/agoric/transactions/';
    }
}
class BldTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Agoric';
        this.family = base_1.CoinFamily.BLD;
        this.explorerUrl = 'https://emerynet.explorer.agoric.net/agoric/tx/';
    }
}
class Sei extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Sei';
        this.family = base_1.CoinFamily.SEI;
        this.explorerUrl = 'https://seitrace.com/tx/';
    }
}
class SeiTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Sei';
        this.family = base_1.CoinFamily.SEI;
        this.explorerUrl = 'https://testnet.seitrace.com/tx/';
    }
}
class Zeta extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Zeta';
        this.family = base_1.CoinFamily.ZETA;
        this.explorerUrl = 'https://explorer.zetachain.com/cosmos/tx/';
    }
}
class ZetaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Zeta';
        this.family = base_1.CoinFamily.ZETA;
        this.explorerUrl = 'https://athens.explorer.zetachain.com/cosmos/tx/';
    }
}
class Injective extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Injective';
        this.family = base_1.CoinFamily.INJECTIVE;
        this.explorerUrl = 'https://www.mintscan.io/injective/tx/';
    }
}
class InjectiveTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'InjectiveTestnet';
        this.family = base_1.CoinFamily.INJECTIVE;
        this.explorerUrl = 'https://testnet.explorer.injective.network/transaction/';
    }
}
class KavaCosmos extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'KavaCosmos';
        this.family = base_1.CoinFamily.KAVACOSMOS;
        this.explorerUrl = 'https://www.mintscan.io/kava/tx/';
        this.addressPrefix = 'kava';
        this.validatorPrefix = 'kavavaloper';
        this.denom = 'ukava';
        this.gasAmount = '5000';
        this.gasLimit = 200000;
        this.validDenoms = ['kava', 'ukava'];
    }
}
class KavaCosmosTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'KavaCosmosTestnet';
        this.family = base_1.CoinFamily.KAVACOSMOS;
        this.explorerUrl = 'https://www.mintscan.io/kava-testnet/tx/';
        this.addressPrefix = 'kava';
        this.validatorPrefix = 'kavavaloper';
        this.denom = 'ukava';
        this.gasAmount = '5000';
        this.gasLimit = 200000;
        this.validDenoms = ['kava', 'ukava'];
    }
}
class Dydx extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Dydx';
        this.family = base_1.CoinFamily.DYDX;
        this.explorerUrl = 'https://www.mintscan.io/dydx/tx/';
        this.addressPrefix = 'dydx';
        this.validatorPrefix = 'dydxvaloper';
        this.denom = 'adydx';
        this.gasAmount = '5000000000000000';
        this.gasLimit = 200000;
        this.validDenoms = ['adydx'];
    }
}
class DydxTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'DydxTestnet';
        this.family = base_1.CoinFamily.DYDX;
        this.explorerUrl = 'https://www.mintscan.io/dydx-testnet/tx/';
        this.addressPrefix = 'dydx';
        this.validatorPrefix = 'dydxvaloper';
        this.denom = 'adv4tnt';
        this.gasAmount = '5000000000000000';
        this.gasLimit = 200000;
        this.validDenoms = ['adv4tnt'];
    }
}
class Ton extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Ton';
        this.family = base_1.CoinFamily.TON;
        this.explorerUrl = 'https://tonscan.org/tx/';
    }
}
class TonTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Ton';
        this.family = base_1.CoinFamily.TON;
        this.explorerUrl = 'https://testnet.tonscan.org/tx/';
    }
}
class Coreum extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'TX';
        this.family = base_1.CoinFamily.COREUM;
        this.explorerUrl = 'https://explorer.tx.org/tx/transactions/';
    }
}
class CoreumTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet TX';
        this.family = base_1.CoinFamily.COREUM;
        this.explorerUrl = 'https://explorer.testnet-1.tx.org/tx/transactions/';
    }
}
class Rune extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Rune';
        this.family = base_1.CoinFamily.THOR;
        this.explorerUrl = 'https://runescan.io/tx/';
    }
}
class RuneTestNet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'RuneTestNet';
        this.family = base_1.CoinFamily.THOR;
        this.explorerUrl = 'https://runescan.io/tx/?network=stagenet';
    }
}
class Baby extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Babylon';
        this.family = base_1.CoinFamily.BABY;
        this.explorerUrl = 'https://babylon.explorers.guru/transaction/';
    }
}
class BabyTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Babylon';
        this.family = base_1.CoinFamily.BABY;
        this.explorerUrl = 'https://testnet.babylon.explorers.guru/transaction/';
    }
}
class Mantra extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Mantra';
        this.family = base_1.CoinFamily.MANTRA;
        this.explorerUrl = 'https://explorer.mantrachain.io/MANTRA-1/tx/';
        this.addressPrefix = 'mantra';
        this.validatorPrefix = 'mantravaloper';
        this.denom = 'uom';
        this.gasAmount = '8000';
        this.gasLimit = 200000;
        this.validDenoms = ['om', 'uom'];
    }
}
class MantraTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Mantra';
        this.family = base_1.CoinFamily.MANTRA;
        this.explorerUrl = 'https://explorer.mantrachain.io/MANTRA-Dukong/tx/';
        this.addressPrefix = 'mantra';
        this.validatorPrefix = 'mantravaloper';
        this.denom = 'uom';
        this.gasAmount = '8000';
        this.gasLimit = 200000;
        this.validDenoms = ['om', 'uom'];
    }
}
class Cronos extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Cronos POS';
        this.family = base_1.CoinFamily.CRONOS;
        this.explorerUrl = 'https://cronos-pos.org/explorer/tx/';
    }
}
class CronosTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Cronos POS';
        this.family = base_1.CoinFamily.CRONOS;
        this.explorerUrl = 'https://cronos-pos.org/explorer/croeseid4/tx/';
    }
}
class FetchAi extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Fetch';
        this.family = base_1.CoinFamily.FETCHAI;
        this.explorerUrl = 'https://www.mintscan.io/fetchai';
    }
}
class FetchAiTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Fetch';
        this.family = base_1.CoinFamily.FETCHAI;
        this.explorerUrl = 'https://explore-eridanus-1.fetch.ai/';
    }
}
class Initia extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Initia';
        this.family = base_1.CoinFamily.INITIA;
        this.explorerUrl = 'https://scan.initia.xyz/interwoven-1/txs/';
    }
}
class InitiaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Initia';
        this.family = base_1.CoinFamily.INITIA;
        this.explorerUrl = 'https://scan.testnet.initia.xyz/initiation-2/txs/';
    }
}
class Asi extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Fetch Native';
        this.family = base_1.CoinFamily.ASI;
        this.explorerUrl = 'https://companion.fetch.ai/fetchhub-4/transactions/';
    }
}
class AsiTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Fetch Native';
        this.family = base_1.CoinFamily.ASI;
        this.explorerUrl = 'https://companion.fetch.ai/dorado-1/transactions/';
    }
}
class Islm extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Haqq';
        this.family = base_1.CoinFamily.ISLM;
        this.explorerUrl = 'https://ping.pub/haqq/tx/';
    }
}
class IslmTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'HaqqTestnet';
        this.family = base_1.CoinFamily.ISLM;
        this.explorerUrl = 'https://testnet.ping.pub/haqq/tx/';
    }
}
class Stx extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Stx';
        this.family = base_1.CoinFamily.STX;
        this.explorerUrl = 'https://explorer.hiro.so/txid/';
        this.sendmanymemoContractAddress = 'SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE';
        this.stakingContractAddress = 'SP000000000000000000002Q6VF78';
    }
}
class StxTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'StxTestnet';
        this.family = base_1.CoinFamily.STX;
        this.explorerUrl = 'https://explorer.hiro.so/txid/?chain=testnet';
        this.sendmanymemoContractAddress = 'ST3F1X4QGV2SM8XD96X45M6RTQXKA1PZJZZCQAB4B';
        this.stakingContractAddress = 'ST000000000000000000002AMW42H';
    }
}
class SUSD extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'SUSD';
        this.family = base_1.CoinFamily.SUSD;
        this.explorerUrl = undefined;
    }
}
class SUSDTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SUSDTestnet';
        this.family = base_1.CoinFamily.SUSD;
        this.explorerUrl = undefined;
    }
}
class FiatTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'FiatTestnet';
        this.family = base_1.CoinFamily.FIAT;
        this.explorerUrl = undefined;
    }
}
class Fiat extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Fiat';
        this.family = base_1.CoinFamily.FIAT;
        this.explorerUrl = undefined;
    }
}
class Bittensor extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Tao';
        this.family = base_1.CoinFamily.TAO;
        this.explorerUrl = 'https://taostats.io/extrinsic/';
    }
}
class BittensorTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Tao';
        this.family = base_1.CoinFamily.TAO;
        this.explorerUrl = 'https://explorer.finney.opentensor.ai/#/';
        this.specName = 'node-subtensor';
        this.genesisHash = '0x8f9cf856bf558a14440e75569c9e58594757048d7b3a84b5d25f6bd978263105';
        this.specVersion = 224;
        this.chainName = 'bittensor';
        this.txVersion = 1;
    }
}
class Trx extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Trx';
        this.family = base_1.CoinFamily.TRX;
        this.accountExplorerUrl = 'https://tronscan.org/#/address/';
        this.explorerUrl = 'https://tronscan.org/#/transaction/';
        this.maxFeeLimit = '5000000000';
        this.contractCallFeeLimit = '100000000';
    }
}
class TrxTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'TrxTestnet';
        this.family = base_1.CoinFamily.TRX;
        this.accountExplorerUrl = 'https://shasta.tronscan.org/#/address/';
        this.explorerUrl = 'https://shasta.tronscan.org/#/transaction/';
        this.maxFeeLimit = '5000000000';
        this.contractCallFeeLimit = '100000000';
    }
}
class Xrp extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Xrp';
        this.family = base_1.CoinFamily.XRP;
        this.explorerUrl = 'https://livenet.xrpl.org/transactions/';
    }
}
class XrpTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'XrpTestnet';
        this.family = base_1.CoinFamily.XRP;
        this.explorerUrl = 'https://test.xrplexplorer.com/explorer/';
    }
}
class Xtz extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Xtz';
        this.family = base_1.CoinFamily.XTZ;
        this.explorerUrl = 'https://tzstats.com/';
        this.accountExplorerUrl = 'https://tzstats.com/';
    }
}
class XtzTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'XtzTestnet';
        this.family = base_1.CoinFamily.XTZ;
        this.explorerUrl = 'https://ghostnet.tzkt.io/';
        this.accountExplorerUrl = 'https://ghostnet.tzkt.io/';
    }
}
class ZCash extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'ZCash';
        this.family = base_1.CoinFamily.ZEC;
        this.utxolibName = 'zcash';
        this.explorerUrl = 'https://mainnet.zcashexplorer.app/transactions/';
    }
}
class ZCashTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ZCashTestnet';
        this.family = base_1.CoinFamily.ZEC;
        this.utxolibName = 'zcashTest';
        this.explorerUrl = 'https://testnet.zcashexplorer.app/transactions/';
    }
}
class Near extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Near';
        this.family = base_1.CoinFamily.NEAR;
        this.explorerUrl = 'https://nearblocks.io/txns/';
        this.feeReserve = '50000000000000000000000';
        this.storageReserve = '2000000000000000000000'; // feeReserve + storageReserve is minimum account balance for a NEAR wallet https://docs.near.org/integrator/faq#is-there-a-minimum-account-balance
    }
}
class NearTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'NearTestnet';
        this.family = base_1.CoinFamily.NEAR;
        this.explorerUrl = 'https://testnet.nearblocks.io/txns/';
        this.feeReserve = '50000000000000000000000';
        this.storageReserve = '2000000000000000000000'; // feeReserve + storageReserve is minimum account balance for a NEAR wallet https://docs.near.org/integrator/faq#is-there-a-minimum-account-balance
    }
}
class Polygon extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Polygon';
        this.family = base_1.CoinFamily.POLYGON;
        this.explorerUrl = 'https://polygonscan.com/tx/';
        this.accountExplorerUrl = 'https://polygonscan.com/address/';
        this.chainId = 137;
        this.forwarderFactoryAddress = '0x29ef46035e9fa3d570c598d3266424ca11413b0c';
        this.forwarderImplementationAddress = '0x5397d0869aba0d55e96d5716d383f6e1d8695ed7';
        this.walletFactoryAddress = '0xa7198f48c58e91f01317e70cd24c5cce475c1555';
        this.walletImplementationAddress = '0xe5dcdc13b628c2df813db1080367e929c1507ca0';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.nativeCoinOperationHashPrefix = 'POLYGON';
        this.tokenOperationHashPrefix = 'POLYGON-ERC20';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class PolygonTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'PolygonTestnet';
        this.family = base_1.CoinFamily.POLYGON;
        this.explorerUrl = 'https://amoy.polygonscan.com/tx/';
        this.accountExplorerUrl = 'https://amoy.polygonscan.com/address/';
        this.chainId = 80002;
        this.forwarderFactoryAddress = '0x23223fd7cfc5f34b20f38ebb9a489e1ff3c084d1';
        this.forwarderImplementationAddress = '0xd4c24325b5ef514fbb1ecc76a9bb4dcb57a8341e';
        this.walletFactoryAddress = '0xe9c16033e01bf9f94bf9f99fc0fefcfb8b124992';
        this.walletImplementationAddress = '0xf8360c213e4f44cce03e48cb9281254de2a7a566';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.nativeCoinOperationHashPrefix = 'POLYGON';
        this.tokenOperationHashPrefix = 'POLYGON-ERC20';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class Optimism extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Optimism';
        this.family = base_1.CoinFamily.OPETH;
        this.explorerUrl = 'https://optimistic.etherscan.io/tx/';
        this.accountExplorerUrl = 'https://optimistic.etherscan.io/address/';
        this.chainId = 10;
        this.nativeCoinOperationHashPrefix = '10';
        this.tokenOperationHashPrefix = '10-ERC20';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.batcherContractAddress = '0xebe27913fcc7510eadf10643a8f86bf5492a9541';
    }
}
class OptimismTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'OptimismTestnet';
        this.family = base_1.CoinFamily.OPETH;
        this.explorerUrl = 'https://sepolia-optimism.etherscan.io/tx/';
        this.accountExplorerUrl = 'https://sepolia-optimism.etherscan.io/address/';
        this.chainId = 11155420;
        this.nativeCoinOperationHashPrefix = '11155420';
        this.tokenOperationHashPrefix = '11155420-ERC20';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.batcherContractAddress = '0xebe27913fcc7510eadf10643a8f86bf5492a9541';
    }
}
class ZkSync extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'ZkSync';
        this.family = base_1.CoinFamily.ZKETH;
        this.explorerUrl = 'https://explorer.zksync.io/tx/';
        this.accountExplorerUrl = 'https://explorer.zksync.io/address/';
        this.chainId = 324;
        this.nativeCoinOperationHashPrefix = '324';
        this.tokenOperationHashPrefix = '324-ERC20';
    }
}
class ZkSyncTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ZkSyncTestnet';
        this.family = base_1.CoinFamily.ZKETH;
        this.explorerUrl = 'https://sepolia.explorer.zksync.io/tx/';
        this.accountExplorerUrl = 'https://sepolia.explorer.zksync.io/address/';
        this.chainId = 300;
        this.nativeCoinOperationHashPrefix = '300';
        this.tokenOperationHashPrefix = '300-ERC20';
        this.forwarderFactoryAddress = '0xdd498702f44c4da08eb9e08d3f015eefe5cb71fc';
        this.forwarderImplementationAddress = '0xbe69cae311191fb45e648ed20847f06fad2dbab4';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x4550e1e7616d3364877fc6c9324938dab678621a';
        this.walletImplementationAddress = '0x92db2759d1dca129a0d9d46877f361be819184c4';
    }
}
class ZkSyncEra extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'ZkSyncEra';
        this.family = base_1.CoinFamily.ZKSYNCERA;
        this.explorerUrl = 'https://explorer.zksync.io/tx/';
        this.accountExplorerUrl = 'https://explorer.zksync.io/address/';
        this.chainId = 324;
        this.nativeCoinOperationHashPrefix = '324';
        this.tokenOperationHashPrefix = '324-ERC20';
    }
}
class ZkSyncEraTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ZkSyncEraTestnet';
        this.family = base_1.CoinFamily.ZKSYNCERA;
        this.explorerUrl = 'https://sepolia.explorer.zksync.io/tx/';
        this.accountExplorerUrl = 'https://sepolia.explorer.zksync.io/address/';
        this.chainId = 300;
        this.nativeCoinOperationHashPrefix = '300';
        this.tokenOperationHashPrefix = '300-ERC20';
        this.forwarderFactoryAddress = '0xdd498702f44c4da08eb9e08d3f015eefe5cb71fc';
        this.forwarderImplementationAddress = '0xbe69cae311191fb45e648ed20847f06fad2dbab4';
        this.walletV4ForwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.walletV4ForwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x4550e1e7616d3364877fc6c9324938dab678621a';
        this.walletImplementationAddress = '0x92db2759d1dca129a0d9d46877f361be819184c4';
    }
}
class Berachain extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Bera';
        this.family = base_1.CoinFamily.BERA;
        this.explorerUrl = 'https://berascan.com/tx/';
        this.accountExplorerUrl = 'https://berascan.com/address/';
        this.chainId = 80094;
        this.nativeCoinOperationHashPrefix = '80094';
        this.tokenOperationHashPrefix = '80094-ERC20';
        this.batcherContractAddress = '0xedf1a0016d9c41d2ad0c275e1ba708361a90c0d1';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class BerachainTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'BeraTestnet';
        this.family = base_1.CoinFamily.BERA;
        this.explorerUrl = 'https://testnet.berascan.com/tx/';
        this.accountExplorerUrl = 'https://testnet.berascan.com/address/';
        this.chainId = 80069;
        this.nativeCoinOperationHashPrefix = '80069';
        this.tokenOperationHashPrefix = '80069-ERC20';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class Oas extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Oas';
        this.family = base_1.CoinFamily.OAS;
        this.explorerUrl = 'https://explorer.oasys.games/tx/';
        this.accountExplorerUrl = 'https://explorer.oasys.games/address/';
        this.chainId = 248;
        this.nativeCoinOperationHashPrefix = '248';
        this.batcherContractAddress = '0xedf1a0016d9c41d2ad0c275e1ba708361a90c0d1';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class OasTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'OasTestnet';
        this.family = base_1.CoinFamily.OAS;
        this.explorerUrl = 'https://explorer.testnet.oasys.games/tx/';
        this.accountExplorerUrl = 'https://explorer.testnet.oasys.games/address/';
        this.chainId = 9372;
        this.nativeCoinOperationHashPrefix = '9372';
        this.batcherContractAddress = '0xedf1a0016d9c41d2ad0c275e1ba708361a90c0d1';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class Coredao extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Coredao';
        this.family = base_1.CoinFamily.COREDAO;
        this.explorerUrl = 'https://scan.coredao.org/tx/';
        this.accountExplorerUrl = 'https://scan.coredao.org/address/';
        this.chainId = 1116;
        this.nativeCoinOperationHashPrefix = '1116';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class CoredaoTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'CoredaoTestnet';
        this.family = base_1.CoinFamily.COREDAO;
        this.explorerUrl = 'https://scan.test2.btcs.network/tx/';
        this.accountExplorerUrl = 'https://scan.test2.btcs.network/address/';
        this.chainId = 1114;
        this.nativeCoinOperationHashPrefix = '1114';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class IP extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Story';
        this.family = base_1.CoinFamily.IP;
        this.explorerUrl = 'https://www.storyscan.io/tx/';
        this.accountExplorerUrl = 'https://www.storyscan.io/address/';
        this.chainId = 1514;
        this.nativeCoinOperationHashPrefix = '1514';
    }
}
class IPTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'StoryTestnet';
        this.family = base_1.CoinFamily.IP;
        this.explorerUrl = 'https://aeneid.explorer.story.foundation/tx/';
        this.accountExplorerUrl = 'https://aeneid.storyscan.io/address/';
        this.chainId = 1315;
        this.nativeCoinOperationHashPrefix = '1315';
    }
}
class ApeChain extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'ApeChain';
        this.family = base_1.CoinFamily.APECHAIN;
        this.explorerUrl = 'https://apescan.io/tx/';
        this.accountExplorerUrl = 'https://apescan.io/address/';
        this.chainId = 33139;
        this.nativeCoinOperationHashPrefix = '33139';
    }
}
class ApeChainTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'ApeChainTestnet';
        this.family = base_1.CoinFamily.APECHAIN;
        this.explorerUrl = 'https://curtis.apescan.io/tx/';
        this.accountExplorerUrl = 'https://curtis.apescan.io/address/';
        this.chainId = 33111;
        this.nativeCoinOperationHashPrefix = '33111';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class Pharos extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Pharos';
        this.family = base_1.CoinFamily.PHRS;
        this.explorerUrl = ''; // TODO: WIN-5787 Add mainnet url when available
        this.accountExplorerUrl = ''; // TODO: WIN-5787 Add mainnet url when available
        this.chainId = 688688; // TODO: WIN-5787 Add mainnet id when available
        this.nativeCoinOperationHashPrefix = '688688';
    }
}
class PharosTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'PharosTestnet';
        this.family = base_1.CoinFamily.PHRS;
        this.explorerUrl = 'https://testnet.pharosscan.xyz/tx/';
        this.accountExplorerUrl = 'https://testnet.pharosscan.xyz/address/';
        this.chainId = 688688;
        this.nativeCoinOperationHashPrefix = '688688';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class Creditcoin extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Creditcoin';
        this.family = base_1.CoinFamily.CTC;
        this.explorerUrl = 'https://creditcoin.blockscout.com/tx/';
        this.accountExplorerUrl = 'https://creditcoin.blockscout.com/address/';
        this.chainId = 102030;
        this.nativeCoinOperationHashPrefix = '102030';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class Plasma extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Plasma';
        this.family = base_1.CoinFamily.XPL;
        this.explorerUrl = 'https://plasmascan.to/tx/';
        this.accountExplorerUrl = 'https://plasmascan.to/address/';
        this.chainId = 9745;
        this.nativeCoinOperationHashPrefix = '9745';
    }
}
class PlasmaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'PlasmaTestnet';
        this.family = base_1.CoinFamily.XPL;
        this.explorerUrl = 'https://testnet.plasmascan.to/tx/';
        this.accountExplorerUrl = 'https://testnet.plasmascan.to/address/';
        this.chainId = 9746;
        this.nativeCoinOperationHashPrefix = '9746';
    }
}
class HypeEVMTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'HyperliquidEVMTestnet';
        this.family = base_1.CoinFamily.HYPEEVM;
        this.explorerUrl = 'https://testnet.purrsec.com/tx/';
        this.accountExplorerUrl = 'https://testnet.purrsec.com/address/';
        this.chainId = 998;
        this.nativeCoinOperationHashPrefix = '998';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class HypeEVM extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'HyperliquidEVM';
        this.family = base_1.CoinFamily.HYPEEVM;
        this.explorerUrl = 'https://hyperevmscan.io/tx/';
        this.accountExplorerUrl = 'https://hyperevmscan.io/address/';
        this.chainId = 999;
        this.nativeCoinOperationHashPrefix = '999';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class KavaEVMTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'KavaEVMTestnet';
        this.family = base_1.CoinFamily.KAVAEVM;
        this.explorerUrl = 'https://testnet.kavascan.com/tx/';
        this.accountExplorerUrl = 'https://testnet.kavascan.com/address/';
        this.chainId = 2221;
        this.nativeCoinOperationHashPrefix = '2221';
    }
}
class KavaEVM extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'KavaEVM';
        this.family = base_1.CoinFamily.KAVAEVM;
        this.explorerUrl = 'https://kavascan.com/tx/';
        this.accountExplorerUrl = 'https://kavascan.com/address/';
        this.chainId = 2222;
        this.nativeCoinOperationHashPrefix = '2222';
    }
}
class LineaETH extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Linea Ethereum Testnet';
        this.family = base_1.CoinFamily.LINEAETH;
        this.explorerUrl = 'https://lineascan.build/tx/';
        this.accountExplorerUrl = 'https://lineascan.build/address/';
        this.chainId = 59144;
        this.nativeCoinOperationHashPrefix = '59144';
    }
}
class LineaETHTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Linea Ethereum Testnet';
        this.family = base_1.CoinFamily.LINEAETH;
        this.explorerUrl = 'https://sepolia.lineascan.build/tx/';
        this.accountExplorerUrl = 'https://sepolia.lineascan.build/address/';
        this.chainId = 59141;
        this.nativeCoinOperationHashPrefix = '59141';
    }
}
class FluentETH extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Fluent Ethereum';
        this.family = base_1.CoinFamily.FLUENTETH;
        this.explorerUrl = 'https://devnet.fluentscan.xyz/tx/';
        this.accountExplorerUrl = 'https://devnet.fluentscan.xyz/address/';
        this.chainId = 20993;
        this.nativeCoinOperationHashPrefix = '20993';
    }
}
class FluentETHTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Fluent Ethereum';
        this.family = base_1.CoinFamily.FLUENTETH;
        this.explorerUrl = 'https://testnet.fluentscan.xyz/tx/';
        this.accountExplorerUrl = 'https://testnet.fluentscan.xyz/address/';
        this.chainId = 20994;
        this.nativeCoinOperationHashPrefix = '20994';
    }
}
class Mantle extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Mantle Native';
        this.family = base_1.CoinFamily.MANTLE;
        this.explorerUrl = 'https://mantlescan.xyz/tx/';
        this.accountExplorerUrl = 'https://mantlescan.xyz/address/';
        this.chainId = 5000;
        this.nativeCoinOperationHashPrefix = '5000';
    }
}
class MantleTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Mantle Native';
        this.family = base_1.CoinFamily.MANTLE;
        this.explorerUrl = 'https://sepolia.mantlescan.xyz/tx/';
        this.accountExplorerUrl = 'https://sepolia.mantlescan.xyz/address/';
        this.chainId = 5003;
        this.nativeCoinOperationHashPrefix = '5003';
    }
}
class JovayETH extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Jovay Ethereum';
        this.family = base_1.CoinFamily.JOVAYETH;
        this.explorerUrl = 'https://explorer.jovay.io/tx/';
        this.accountExplorerUrl = 'https://explorer.jovay.io/address/';
        this.chainId = 5734951;
        this.nativeCoinOperationHashPrefix = '5734951';
    }
}
class JovayETHTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Jovay Ethereum';
        this.family = base_1.CoinFamily.JOVAYETH;
        this.explorerUrl = 'https://sepolia-explorer.jovay.io/tx/';
        this.accountExplorerUrl = 'https://sepolia-explorer.jovay.io/address/';
        this.chainId = 2019775;
        this.nativeCoinOperationHashPrefix = '2019775';
    }
}
class Unieth extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Unichain';
        this.family = base_1.CoinFamily.UNIETH;
        this.explorerUrl = 'https://uniscan.xyz/tx/';
        this.accountExplorerUrl = 'https://uniscan.xyz/address/';
        this.chainId = 130;
        this.nativeCoinOperationHashPrefix = '130';
    }
}
class UniethTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Unichain';
        this.family = base_1.CoinFamily.UNIETH;
        this.explorerUrl = 'https://sepolia.uniscan.xyz/tx/';
        this.accountExplorerUrl = 'https://sepolia.uniscan.xyz/address/';
        this.chainId = 1301;
        this.nativeCoinOperationHashPrefix = '1301';
    }
}
class Hppeth extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'House Party Protocol';
        this.family = base_1.CoinFamily.HPPETH;
        this.explorerUrl = 'https://explorer.hpp.io/tx/';
        this.accountExplorerUrl = 'https://explorer.hpp.io/address/';
        this.chainId = 190415;
        this.nativeCoinOperationHashPrefix = '190415';
    }
}
class HppethTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet House Party Protocol';
        this.family = base_1.CoinFamily.HPPETH;
        this.explorerUrl = 'https://sepolia-explorer.hpp.io/tx/';
        this.accountExplorerUrl = 'https://sepolia-explorer.hpp.io/address/';
        this.chainId = 181228;
        this.nativeCoinOperationHashPrefix = '181228';
    }
}
class HoodethTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Robinhood Chain';
        this.family = base_1.CoinFamily.HOODETH;
        this.explorerUrl = 'https://explorer.testnet.chain.robinhood.com/tx/';
        this.accountExplorerUrl = 'https://explorer.testnet.chain.robinhood.com/address/';
        this.chainId = 46630;
        this.nativeCoinOperationHashPrefix = '46630';
    }
}
class Hoodeth extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Robinhood Chain';
        this.family = base_1.CoinFamily.HOODETH;
        this.explorerUrl = 'https://explorer.testnet.chain.robinhood.com/tx/'; // TODO: update to mainnet explorer when available
        this.accountExplorerUrl = 'https://explorer.testnet.chain.robinhood.com/address/'; // TODO: update to mainnet explorer when available
        this.chainId = 46630; // TODO: update to mainnet chain id when available
        this.nativeCoinOperationHashPrefix = '46630'; // TODO: update to mainnet chain id when available
    }
}
class HumanityProtocol extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Humanity Protocol';
        this.family = base_1.CoinFamily.H;
        this.explorerUrl = 'https://explorer.humanity.org/mainnet/txns/';
        this.accountExplorerUrl = 'https://explorer.humanity.org/mainnet/accounts/';
        this.chainId = 6985385;
        this.nativeCoinOperationHashPrefix = '6985385';
    }
}
class HumanityProtocolTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Humanity Protocol';
        this.family = base_1.CoinFamily.H;
        this.explorerUrl = 'https://explorer.humanity.org/testnet/txns/';
        this.accountExplorerUrl = 'https://explorer.humanity.org/testnet/accounts/';
        this.chainId = 7080969;
        this.nativeCoinOperationHashPrefix = '7080969';
    }
}
class Xlayer extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'X Layer';
        this.family = base_1.CoinFamily.OKBXLAYER;
        this.explorerUrl = 'https://www.oklink.com/x-layer/tx/';
        this.accountExplorerUrl = 'https://www.oklink.com/x-layer/address/';
        this.chainId = 196;
        this.nativeCoinOperationHashPrefix = '196';
    }
}
class XlayerTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet X Layer';
        this.family = base_1.CoinFamily.OKBXLAYER;
        this.explorerUrl = 'https://www.oklink.com/x-layer-testnet/tx/';
        this.accountExplorerUrl = 'https://www.oklink.com/x-layer-testnet/address/';
        this.chainId = 1952;
        this.nativeCoinOperationHashPrefix = '1952';
    }
}
class Morph extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Morph Mainnet';
        this.family = base_1.CoinFamily.MORPH;
        this.explorerUrl = 'https://explorer-api-hoodi.morphl2.io/tx/'; // TODO: WIN-8134 add mainnet url when available
        this.accountExplorerUrl = 'https://explorer-api-hoodi.morphl2.io/address/'; // TODO: WIN-8134 add mainnet account explorer url when available
        this.chainId = 2818;
        this.nativeCoinOperationHashPrefix = '2818';
    }
}
class MorphTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Morph Testnet';
        this.family = base_1.CoinFamily.MORPH;
        this.explorerUrl = 'https://explorer-hoodi.morph.network/tx/';
        this.accountExplorerUrl = 'https://explorer-hoodi.morph.network/address/';
        this.chainId = 2910;
        this.nativeCoinOperationHashPrefix = '2910';
    }
}
class MorphETH extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Morph Ethereum Mainnet';
        this.family = base_1.CoinFamily.MORPHETH;
        this.explorerUrl = 'https://explorer.morph.network/tx/';
        this.accountExplorerUrl = 'https://explorer.morph.network/address/';
        this.chainId = 2818;
        this.nativeCoinOperationHashPrefix = '2818';
    }
}
class MorphETHTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Morph Ethereum Testnet';
        this.family = base_1.CoinFamily.MORPHETH;
        this.explorerUrl = 'https://explorer-hoodi.morph.network/tx/';
        this.accountExplorerUrl = 'https://explorer-hoodi.morph.network/address/';
        this.chainId = 2910;
        this.nativeCoinOperationHashPrefix = '2910';
    }
}
class CreditcoinTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'CreditcoinTestnet';
        this.family = base_1.CoinFamily.CTC;
        this.explorerUrl = 'https://creditcoin-testnet.blockscout.com/tx/';
        this.accountExplorerUrl = 'https://creditcoin-testnet.blockscout.com/address/';
        this.chainId = 102031;
        this.nativeCoinOperationHashPrefix = '102031';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class Sonic extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Sonic';
        this.family = base_1.CoinFamily.SONIC;
        this.explorerUrl = 'https://sonicscan.org/tx/';
        this.accountExplorerUrl = 'https://sonicscan.org/address/';
        this.chainId = 146;
        this.nativeCoinOperationHashPrefix = '146';
    }
}
class SonicTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SonicTestnet';
        this.family = base_1.CoinFamily.SONIC;
        this.explorerUrl = 'https://testnet.sonicscan.org/tx/';
        this.accountExplorerUrl = 'https://testnet.sonicscan.org/address/';
        this.chainId = 14601;
        this.nativeCoinOperationHashPrefix = '14601';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Usdt0 extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'USDT0';
        this.family = base_1.CoinFamily.USDT0;
        this.explorerUrl = 'https://stablescan.xyz/tx/';
        this.accountExplorerUrl = 'https://stablescan.xyz/address/';
        this.chainId = 988;
        this.nativeCoinOperationHashPrefix = '988';
    }
}
class Usdt0Testnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet USDT0';
        this.family = base_1.CoinFamily.USDT0;
        this.explorerUrl = 'https://testnet.stablescan.xyz/tx/';
        this.accountExplorerUrl = 'https://testnet.stablescan.xyz/address/';
        this.chainId = 2201;
        this.nativeCoinOperationHashPrefix = '2201';
    }
}
class Kaia extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Kaia';
        this.family = base_1.CoinFamily.KAIA;
        this.explorerUrl = 'https://kaiascan.io/tx/';
        this.accountExplorerUrl = 'https://kaiascan.io/address/';
        this.chainId = 8217;
        this.nativeCoinOperationHashPrefix = '8217';
    }
}
class KaiaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'KaiaTestnet';
        this.family = base_1.CoinFamily.KAIA;
        this.explorerUrl = 'https://kairos.kaiascan.io/tx/';
        this.accountExplorerUrl = 'https://kairos.kaiascan.io/address/';
        this.chainId = 1001;
        this.nativeCoinOperationHashPrefix = '1001';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Irys extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Irys';
        this.family = base_1.CoinFamily.IRYS;
        this.explorerUrl = 'https://evm-explorer.irys.xyz/tx/';
        this.accountExplorerUrl = 'https://evm-explorer.irys.xyz/address/';
        this.chainId = 3282;
        this.nativeCoinOperationHashPrefix = '3282';
    }
}
class IrysTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'IrysTestnet';
        this.family = base_1.CoinFamily.IRYS;
        this.explorerUrl = 'https://testnet-explorer.irys.xyz/tx/';
        this.accountExplorerUrl = 'https://testnet-explorer.irys.xyz/address/';
        this.chainId = 1270;
        this.nativeCoinOperationHashPrefix = '1270';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Og extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Zero Gravity';
        this.family = base_1.CoinFamily.OG;
        this.explorerUrl = 'https://chainscan.0g.ai/tx/';
        this.accountExplorerUrl = 'https://chainscan.0g.ai/address/';
        this.chainId = 16661;
        this.nativeCoinOperationHashPrefix = '16661';
    }
}
class OgTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Zero Gravity Testnet';
        this.family = base_1.CoinFamily.OG;
        this.explorerUrl = 'https://chainscan-galileo.0g.ai/tx/';
        this.accountExplorerUrl = 'https://chainscan-galileo.0g.ai/address/';
        this.chainId = 16602;
        this.nativeCoinOperationHashPrefix = '16602';
    }
}
class Xdc extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'XdcChain';
        this.family = base_1.CoinFamily.XDC;
        this.explorerUrl = 'https://xdcscan.com/tx/';
        this.accountExplorerUrl = 'https://xdcscan.com/address/';
        this.chainId = 50;
        this.nativeCoinOperationHashPrefix = '50';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class XdcTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'XdcChainTestnet';
        this.family = base_1.CoinFamily.XDC;
        this.explorerUrl = 'https://testnet.xdcscan.com/tx/';
        this.accountExplorerUrl = 'https://testnet.xdcscan.com/address/';
        this.chainId = 51;
        this.nativeCoinOperationHashPrefix = '51';
        this.batcherContractAddress = '0xebe27913fcc7510eadf10643a8f86bf5492a9541';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class Wemix extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'WemixChain';
        this.family = base_1.CoinFamily.WEMIX;
        this.explorerUrl = 'https://explorer.wemix.com/tx/';
        this.accountExplorerUrl = 'https://explorer.wemix.com/address/';
        this.chainId = 1111;
        this.nativeCoinOperationHashPrefix = '1111';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class WemixTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'WemixChainTestnet';
        this.family = base_1.CoinFamily.WEMIX;
        this.explorerUrl = 'https://explorer.test.wemix.com/tx/';
        this.accountExplorerUrl = 'https://explorer.test.wemix.com/address/';
        this.chainId = 1112;
        this.nativeCoinOperationHashPrefix = '1112';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class MonadTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'MonadTestnet';
        this.family = base_1.CoinFamily.MON;
        this.explorerUrl = 'https://testnet.monadexplorer.com/tx/';
        this.accountExplorerUrl = 'https://testnet.monadexplorer.com/address/';
        this.chainId = 10143;
        this.nativeCoinOperationHashPrefix = '10143';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Monad extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Monad';
        this.family = base_1.CoinFamily.MON;
        this.explorerUrl = 'https://mainnet-beta.monvision.io/tx/';
        this.chainId = 143;
        this.nativeCoinOperationHashPrefix = '143';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class World extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Worldchain';
        this.family = base_1.CoinFamily.WORLD;
        this.explorerUrl = 'https://worldscan.org/tx/';
        this.accountExplorerUrl = 'https://worldscan.org/address/';
        this.chainId = 480;
        this.nativeCoinOperationHashPrefix = '480';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class WorldTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'WorldchainTestnet';
        this.family = base_1.CoinFamily.WORLD;
        this.explorerUrl = 'https://sepolia.worldscan.org/tx/';
        this.accountExplorerUrl = 'https://sepolia.worldscan.org/address/';
        this.chainId = 4801;
        this.nativeCoinOperationHashPrefix = '4801';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Soneium extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Soneium';
        this.family = base_1.CoinFamily.SONEIUM;
        this.explorerUrl = 'https://soneium.blockscout.com/tx/';
        this.accountExplorerUrl = 'https://soneium.blockscout.com/address/';
        this.chainId = 1868;
        this.nativeCoinOperationHashPrefix = '1868';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class SoneiumTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SoneiumTestnet';
        this.family = base_1.CoinFamily.SONEIUM;
        this.explorerUrl = 'https://soneium-minato.blockscout.com/tx/';
        this.accountExplorerUrl = 'https://soneium-minato.blockscout.com/address/';
        this.chainId = 1946;
        this.nativeCoinOperationHashPrefix = '1946';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class SeiEvm extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Seievm';
        this.family = base_1.CoinFamily.SEIEVM;
        this.explorerUrl = 'https://seitrace.com/tx/';
        this.accountExplorerUrl = 'https://seitrace.com/address/';
        this.chainId = 1329;
        this.nativeCoinOperationHashPrefix = '1329';
    }
}
class SeiEvmTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SeiEvmTestnet';
        this.family = base_1.CoinFamily.SEIEVM;
        this.explorerUrl = 'https://testnet.seistream.app/transactions/';
        this.accountExplorerUrl = 'https://testnet.seistream.app/account/';
        this.chainId = 1328;
        this.nativeCoinOperationHashPrefix = '1328';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class SomniaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SomniaTestnet';
        this.family = base_1.CoinFamily.STT;
        this.explorerUrl = 'https://shannon-explorer.somnia.network/tx/';
        this.accountExplorerUrl = 'https://shannon-explorer.somnia.network/address/';
        this.chainId = 50312;
        this.nativeCoinOperationHashPrefix = '50312';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
    }
}
class Somi extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Somnia';
        this.family = base_1.CoinFamily.SOMI;
        this.explorerUrl = 'https://mainnet.somnia.w3us.site/tx/';
        this.chainId = 5031;
        this.nativeCoinOperationHashPrefix = '5031';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.batcherContractAddress = '0x3e1e5d78e44f15593b3b61ed278f12c27f0ff33e';
    }
}
class FlareP extends Mainnet {
    constructor() {
        super(...arguments);
        this.assetId = 'Flare';
        this.name = 'FlareP';
        this.family = base_1.CoinFamily.FLRP;
        this.explorerUrl = 'https://flarescan.com/blockchain/pvm/transactions/';
        this.accountExplorerUrl = 'https://flarescan.com/blockchain/pvm/address/';
        this.blockchainID = '11111111111111111111111111111111LpoYY';
        this.cChainBlockchainID = '2q9e4r6Mu3U68nU1fYjgbR6JvwrRx36CohpAX5UQxse55x1Q5';
        this.xChainBlockchainID = 'FJuSwZuP85eyBpuBrKECnpPedGyXoDy2hP9q4JD8qBTZGxJ';
        this.networkID = 14;
        this.hrp = 'flare';
        this.alias = 'P';
        this.vm = 'platformvm';
        this.minImportToPFee = '1261000'; // 0.1261 FLR
        this.txFee = '200000'; // FLR P-chain import requires higher fee than base txFee
        this.baseTxFee = '1000000';
        this.maxImportFee = '10000000'; // defaults
        this.createAssetTxFee = '1000000';
        this.createSubnetTx = '100000000'; // defaults
        this.transformSubnetTxFee = '100000000';
        this.createChainTx = '100000000'; // defaults
        this.creationTxFee = '10000000'; // defaults
        this.minConsumption = '0.1';
        this.maxConsumption = '0.12';
        this.maxSupply = '103688266974000000000'; // 103B tokens
        this.minStake = '1000000000000000'; // 1M FLR
        this.minStakeDuration = '5256000'; // 2 months
        this.maxStakeDuration = '31536000'; // 1 year
        this.minDelegationStake = '50000000000000'; // 50000 FLR
        this.minDelegationFee = '0';
        this.addPrimaryNetworkValidatorFee = '0';
        this.addSubnetValidatorFee = '1000000';
        this.addSubnetDelegatorFee = '1000000';
    }
}
exports.FlareP = FlareP;
class FlarePTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'FlarePTestnet';
        this.family = base_1.CoinFamily.FLRP;
        this.explorerUrl = 'https://coston2.testnet.flarescan.com/blockchain/pvm/transactions';
        this.accountExplorerUrl = 'https://coston2.testnet.flarescan.com/blockchain/pvm/address/';
        this.flarePublicUrl = 'https://coston2.testnet.flare.network';
        this.blockchainID = '11111111111111111111111111111111LpoYY';
        this.cChainBlockchainID = 'vE8M98mEQH6wk56sStD1ML8HApTgSqfJZLk9gQ3Fsd4i6m3Bi';
        this.xChainBlockchainID = 'FJuSwZuP85eyBpuBrKECnpPedGyXoDy2hP9q4JD8qBTZGxYJ';
        this.networkID = 114;
        this.hrp = 'costwo';
        this.alias = 'P';
        this.assetId = 'fxMAKpBQQpFedrUhWMsDYfCUJxdUw4mneTczKBzNg3rc2JUub';
        this.vm = 'platformvm';
        this.minImportToPFee = '1261000'; // 0.1261 FLR
        this.txFee = '200000'; // FLR P-chain import requires higher fee than base txFee
        this.baseTxFee = '1000000';
        this.maxImportFee = '10000000'; // defaults
        this.createAssetTxFee = '1000000';
        this.createSubnetTx = '100000000'; // defaults
        this.transformSubnetTxFee = '100000000';
        this.createChainTx = '100000000'; // defaults
        this.creationTxFee = '10000000'; // defaults
        this.minConsumption = '0.1';
        this.maxConsumption = '0.12';
        this.maxSupply = '79000000000000000000000'; // 79 trillion tokens
        this.minStake = '1000000000000000'; // 1M FLR
        this.minStakeDuration = '5256000'; // 2 months
        this.maxStakeDuration = '31536000'; // 1 year
        this.minDelegationStake = '50000000000000'; // 50000 FLR
        this.minDelegationFee = '0';
        this.addPrimaryNetworkValidatorFee = '0';
        this.addSubnetValidatorFee = '1000000';
        this.addSubnetDelegatorFee = '1000000';
    }
}
exports.FlarePTestnet = FlarePTestnet;
class Flare extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Flarechain';
        this.family = base_1.CoinFamily.FLR;
        this.explorerUrl = 'https://flare-explorer.flare.network/tx/';
        this.accountExplorerUrl = 'https://flare-explorer.flare.network/address/';
        this.flarePublicUrl = 'https://flare-explorer.flare.network';
        this.chainId = 14;
        this.nativeCoinOperationHashPrefix = '14';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
exports.Flare = Flare;
class FlareTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'FlarechainTestnet';
        this.family = base_1.CoinFamily.FLR;
        this.explorerUrl = 'https://coston2-explorer.flare.network/tx/';
        this.accountExplorerUrl = 'https://coston2-explorer.flare.network/address/';
        this.chainId = 114;
        this.nativeCoinOperationHashPrefix = '114';
        this.walletFactoryAddress = '0x809ee567e413543af1caebcdb247f6a67eafc8dd';
        this.walletImplementationAddress = '0x944fef03af368414f29dc31a72061b8d64f568d2';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
exports.FlareTestnet = FlareTestnet;
class Songbird extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Songbirdchain';
        this.family = base_1.CoinFamily.SGB;
        this.explorerUrl = 'https://songbird-explorer.flare.network/tx/';
        this.accountExplorerUrl = 'https://songbird-explorer.flare.network/address/';
        this.chainId = 19;
        this.nativeCoinOperationHashPrefix = '19';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class SongbirdTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'SongbirdchainTestnet';
        this.family = base_1.CoinFamily.SGB;
        this.explorerUrl = 'https://coston-explorer.flare.network/tx/';
        this.accountExplorerUrl = 'https://coston-explorer.flare.network/address/';
        this.chainId = 16;
        this.nativeCoinOperationHashPrefix = '16';
        this.batcherContractAddress = '0xc5d7a3b4c490c6242fb28f20e5fe33cd983372a3';
        this.forwarderFactoryAddress = '0x37996e762fa8b671869740c79eb33f625b3bf92a';
        this.forwarderImplementationAddress = '0xd5fe1c1f216b775dfd30638fa7164d41321ef79b';
    }
}
class BaseChainTestnet extends Testnet {
    constructor() {
        super(...arguments);
        //BaseETH testnet
        this.name = 'BaseChainTestnet';
        this.family = base_1.CoinFamily.BASEETH;
        this.explorerUrl = 'https://sepolia.basescan.org/tx/';
        this.accountExplorerUrl = 'https://sepolia.basescan.org/address/';
        this.chainId = 84532;
        this.nativeCoinOperationHashPrefix = '84532';
        this.tokenOperationHashPrefix = '84532-ERC20';
        this.forwarderFactoryAddress = '0xdd498702f44c4da08eb9e08d3f015eefe5cb71fc';
        this.forwarderImplementationAddress = '0xbe69cae311191fb45e648ed20847f06fad2dbab4';
        this.walletFactoryAddress = '0x4550e1e7616d3364877fc6c9324938dab678621a';
        this.walletImplementationAddress = '0x92db2759d1dca129a0d9d46877f361be819184c4';
    }
}
class BaseChain extends Mainnet {
    constructor() {
        super(...arguments);
        //BaseETH
        this.name = 'BaseChain';
        this.family = base_1.CoinFamily.BASEETH;
        this.explorerUrl = 'https://basescan.org/tx/';
        this.accountExplorerUrl = 'https://basescan.org/address/';
        this.chainId = 8453;
        this.nativeCoinOperationHashPrefix = '8453';
        this.tokenOperationHashPrefix = '8453-ERC20';
        this.forwarderFactoryAddress = '0xdd498702f44c4da08eb9e08d3f015eefe5cb71fc';
        this.forwarderImplementationAddress = '0xbe69cae311191fb45e648ed20847f06fad2dbab4';
        this.walletFactoryAddress = '0x4550e1e7616d3364877fc6c9324938dab678621a';
        this.walletImplementationAddress = '0x92db2759d1dca129a0d9d46877f361be819184c4';
    }
}
class Polymesh extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Polymesh';
        this.family = base_1.CoinFamily.POLYX;
        this.explorerUrl = 'https://polymesh.subscan.io/extrinsic/';
        this.specName = 'polymesh_mainnet';
        this.genesisHash = '0x6fbd74e5e1d0a61d52ccfe9d4adaed16dd3a7caa37c6bc4d0c2fa12e8b2f4063';
        this.specVersion = 7002000;
        this.chainName = 'Polymesh Mainnet';
        this.txVersion = 7;
    }
}
class PolymeshTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'PolymeshTestnet';
        this.family = base_1.CoinFamily.POLYX;
        this.explorerUrl = 'https://polymesh-testnet.subscan.io/extrinsic/';
        this.specName = 'polymesh_testnet';
        this.genesisHash = '0x2ace05e703aa50b48c0ccccfc8b424f7aab9a1e2c424ed12e45d20b1e8ffd0d6';
        this.specVersion = 7002000;
        this.chainName = 'Polymesh Testnet';
        this.txVersion = 7;
    }
}
class Vet extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'VeChain';
        this.family = base_1.CoinFamily.VET;
        this.explorerUrl = 'https://explore.vechain.org/transactions/';
        this.accountExplorerUrl = 'https://explore.vechain.org/accounts/';
        this.chainId = 100009;
        this.forwarderFactoryAddress = '0x65343e18c376d2fc8c3cf10cd146d63e2e0dc9ef';
        this.forwarderImplementationAddress = '0x62de34c87f847d385af07f6c25dbd97b1fffefc0';
    }
}
class VetTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'VeChainTestnet';
        this.family = base_1.CoinFamily.VET;
        this.explorerUrl = 'https://explore-testnet.vechain.org/transactions/';
        this.accountExplorerUrl = 'https://explore-testnet.vechain.org/accounts/';
        this.chainId = 100010;
        this.forwarderFactoryAddress = '0x65343e18c376d2fc8c3cf10cd146d63e2e0dc9ef';
        this.forwarderImplementationAddress = '0x62de34c87f847d385af07f6c25dbd97b1fffefc0';
    }
}
class Iota extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Iota';
        this.family = base_1.CoinFamily.IOTA;
        this.explorerUrl = 'https://explorer.iota.org/txblock/?network=mainnet';
        this.accountExplorerUrl = 'https://explorer.iota.org/address/?network=mainnet';
    }
}
class IotaTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Iota';
        this.family = base_1.CoinFamily.IOTA;
        this.explorerUrl = 'https://explorer.iota.org/txblock/?network=testnet';
        this.accountExplorerUrl = 'https://explorer.iota.org/address/?network=testnet';
    }
}
class Flow extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Flow';
        this.family = base_1.CoinFamily.FLOW;
        this.explorerUrl = 'https://evm.flowscan.io/tx/';
        this.accountExplorerUrl = 'https://evm.flowscan.io/address/';
        this.chainId = 747;
        this.nativeCoinOperationHashPrefix = '747';
    }
}
class FlowTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'FlowTestnet';
        this.family = base_1.CoinFamily.FLOW;
        this.explorerUrl = 'https://evm-testnet.flowscan.io/tx/';
        this.accountExplorerUrl = 'https://evm-testnet.flowscan.io/address/';
        this.chainId = 545;
        this.nativeCoinOperationHashPrefix = '545';
    }
}
class MegaETH extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'MegaETH';
        this.family = base_1.CoinFamily.MEGAETH;
        this.explorerUrl = 'https://megaeth.blockscout.com/tx/';
        this.accountExplorerUrl = 'https://megaeth.blockscout.com/address/';
        this.chainId = 4326;
        this.nativeCoinOperationHashPrefix = '4326';
    }
}
class MegaETHTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'MegaETHTestnet';
        this.family = base_1.CoinFamily.MEGAETH;
        this.explorerUrl = 'https://www.megaexplorer.xyz/tx/';
        this.accountExplorerUrl = 'https://www.megaexplorer.xyz/address/';
        this.chainId = 6343;
        this.nativeCoinOperationHashPrefix = '6343';
    }
}
class Plume extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Plume';
        this.family = base_1.CoinFamily.PLUME;
        this.explorerUrl = 'https://explorer.plume.org/tx/';
        this.accountExplorerUrl = 'https://explorer.plume.org/address/';
        this.chainId = 98866;
        this.nativeCoinOperationHashPrefix = '98866';
    }
}
class HederaEVMTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Hedera EVM';
        this.family = base_1.CoinFamily.HBAREVM;
        this.explorerUrl = 'https://hashscan.io/testnet/transaction/';
        this.accountExplorerUrl = 'https://hashscan.io/testnet/account/';
        this.chainId = 296;
        this.nativeCoinOperationHashPrefix = '296';
    }
}
class HederaEVM extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Hedera EVM';
        this.family = base_1.CoinFamily.HBAREVM;
        this.explorerUrl = 'https://hashscan.io/mainnet/transaction/';
        this.accountExplorerUrl = 'https://hashscan.io/mainnet/account/';
        this.chainId = 295;
        this.nativeCoinOperationHashPrefix = '295';
    }
}
class HemiEth extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Hemi Ethereum';
        this.family = base_1.CoinFamily.HEMIETH;
        this.explorerUrl = 'https://explorer.hemi.xyz/tx/';
        this.accountExplorerUrl = 'https://explorer.hemi.xyz/address/';
        this.chainId = 43111;
        this.nativeCoinOperationHashPrefix = '43111';
    }
}
class HemiEthTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Testnet Hemi Ethereum';
        this.family = base_1.CoinFamily.HEMIETH;
        this.explorerUrl = 'https://testnet.explorer.hemi.xyz/tx/';
        this.accountExplorerUrl = 'https://testnet.explorer.hemi.xyz/address/';
        this.chainId = 743111;
        this.nativeCoinOperationHashPrefix = '743111';
    }
}
class PlumeTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'PlumeTestnet';
        this.family = base_1.CoinFamily.PLUME;
        this.explorerUrl = 'https://testnet-explorer.plume.org/tx/';
        this.accountExplorerUrl = 'https://testnet-explorer.plume.org/address/';
        this.chainId = 98867;
        this.nativeCoinOperationHashPrefix = '98867';
    }
}
class Canton extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Canton';
        this.family = base_1.CoinFamily.CANTON;
        this.explorerUrl = 'https://lighthouse.cantonloop.com/transactions/';
    }
}
class CantonTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'CantonTestnet';
        this.family = base_1.CoinFamily.CANTON;
        this.explorerUrl = 'https://lighthouse.devnet.cantonloop.com/transactions/';
    }
}
class Dogeos extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Doge OS';
        this.family = base_1.CoinFamily.DOGEOS;
        this.explorerUrl = 'https://blockscout.testnet.dogeos.com/tx/'; //TODO: WIN-8082 => add mainnet url when available
        this.accountExplorerUrl = "'https://blockscout.testnet.dogeos.com/address/"; //TODO: WIN-8082 => add mainnet url when available
        this.chainId = 1234567; ////TODO: WIN-8082 => add mainnet chainId when available
        this.nativeCoinOperationHashPrefix = '1234567'; //TODO: WIN-8082 => add when mainnet details available
    }
}
class DogeosTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Doge OS Testnet';
        this.family = base_1.CoinFamily.DOGEOS;
        this.explorerUrl = 'https://blockscout.testnet.dogeos.com/tx/';
        this.accountExplorerUrl = 'https://blockscout.testnet.dogeos.com/address/';
        this.chainId = 6281971;
        this.nativeCoinOperationHashPrefix = '6281971';
    }
}
class ArcUSDCTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'USDC (ARC) Testnet';
        this.family = base_1.CoinFamily.ARCUSDC;
        this.explorerUrl = 'https://testnet.arcscan.app/tx/';
        this.accountExplorerUrl = 'https://testnet.arcscan.app/address/';
        this.chainId = 5042002;
        this.nativeCoinOperationHashPrefix = '5042002';
    }
}
class ArcUSDC extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'USDC (ARC)';
        this.family = base_1.CoinFamily.ARCUSDC;
        this.explorerUrl = 'https://testnet.arcscan.app/tx/'; // WIN-8844 change to prod explorer url
        this.accountExplorerUrl = 'https://testnet.arcscan.app/address/'; // change to prod explorer url when available
        this.chainId = 5042002; // change to prod chain id when available
        this.nativeCoinOperationHashPrefix = '5042002'; // change to prod native coin operation hash prefix when available
    }
}
class Tempo extends Mainnet {
    constructor() {
        super(...arguments);
        this.name = 'Tempo';
        this.family = base_1.CoinFamily.TEMPO;
        this.explorerUrl = 'https://explore.mainnet.tempo.xyz/tx/';
        this.accountExplorerUrl = 'https://explore.mainnet.tempo.xyz/address/';
        this.chainId = 4217;
        this.nativeCoinOperationHashPrefix = 'TEMPO';
        this.tokenOperationHashPrefix = 'TEMPO-TIP20';
    }
}
class TempoTestnet extends Testnet {
    constructor() {
        super(...arguments);
        this.name = 'Tempo Testnet';
        this.family = base_1.CoinFamily.TEMPO;
        this.explorerUrl = 'https://explore.tempo.xyz/tx/';
        this.accountExplorerUrl = 'https://explore.tempo.xyz/address/';
        this.chainId = 42431;
        this.nativeCoinOperationHashPrefix = '42431';
        this.tokenOperationHashPrefix = '42431';
    }
}
exports.Networks = {
    main: {
        ada: Object.freeze(new Ada()),
        algorand: Object.freeze(new Algorand()),
        apechain: Object.freeze(new ApeChain()),
        apt: Object.freeze(new Apt()),
        arbitrum: Object.freeze(new Arbitrum()),
        arcusdc: Object.freeze(new ArcUSDC()),
        asi: Object.freeze(new Asi()),
        atom: Object.freeze(new Atom()),
        avalancheC: Object.freeze(new AvalancheC()),
        avalancheP: Object.freeze(new AvalancheP()),
        baby: Object.freeze(new Baby()),
        basechain: Object.freeze(new BaseChain()),
        bitcoin: Object.freeze(new Bitcoin()),
        bitcoinCash: Object.freeze(new BitcoinCash()),
        bitcoinGold: Object.freeze(new BitcoinGold()),
        bitcoinSV: Object.freeze(new BitcoinSV()),
        bera: Object.freeze(new Berachain()),
        bld: Object.freeze(new Bld()),
        bsc: Object.freeze(new BinanceSmartChain()),
        canton: Object.freeze(new Canton()),
        casper: Object.freeze(new Casper()),
        celo: Object.freeze(new Celo()),
        chiliz: Object.freeze(new Chiliz()),
        codexeth: Object.freeze(new CodexEth()),
        coredao: Object.freeze(new Coredao()),
        coreum: Object.freeze(new Coreum()),
        cronos: Object.freeze(new Cronos()),
        dash: Object.freeze(new Dash()),
        dogecoin: Object.freeze(new Dogecoin()),
        dogeos: Object.freeze(new Dogeos()),
        dot: Object.freeze(new PolkadotAssetHub()),
        dydx: Object.freeze(new Dydx()),
        eCash: Object.freeze(new ECash()),
        eos: Object.freeze(new Eos()),
        ethereum: Object.freeze(new Ethereum()),
        ethereum2: Object.freeze(new Ethereum2()),
        ethereumClassic: Object.freeze(new EthereumClassic()),
        ethereumW: Object.freeze(new EthereumW()),
        fiat: Object.freeze(new Fiat()),
        fetchai: Object.freeze(new FetchAi()),
        flow: Object.freeze(new Flow()),
        flr: Object.freeze(new Flare()),
        flrP: Object.freeze(new FlareP()),
        fluenteth: Object.freeze(new FluentETH()),
        h: Object.freeze(new HumanityProtocol()),
        hash: Object.freeze(new Hash()),
        hoodeth: Object.freeze(new Hoodeth()),
        hedera: Object.freeze(new Hedera()),
        hederaEVM: Object.freeze(new HederaEVM()),
        hemieth: Object.freeze(new HemiEth()),
        hppeth: Object.freeze(new Hppeth()),
        icp: Object.freeze(new Icp()),
        ip: Object.freeze(new IP()),
        initia: Object.freeze(new Initia()),
        iota: Object.freeze(new Iota()),
        injective: Object.freeze(new Injective()),
        irys: Object.freeze(new Irys()),
        islm: Object.freeze(new Islm()),
        jovayeth: Object.freeze(new JovayETH()),
        kaia: Object.freeze(new Kaia()),
        kavacosmos: Object.freeze(new KavaCosmos()),
        kavaevm: Object.freeze(new KavaEVM()),
        lnbtc: Object.freeze(new LightningBitcoin()),
        litecoin: Object.freeze(new Litecoin()),
        megaeth: Object.freeze(new MegaETH()),
        mon: Object.freeze(new Monad()),
        mantle: Object.freeze(new Mantle()),
        mantra: Object.freeze(new Mantra()),
        plume: Object.freeze(new Plume()),
        polygon: Object.freeze(new Polygon()),
        polyx: Object.freeze(new Polymesh()),
        phrs: Object.freeze(new Pharos()),
        ctc: Object.freeze(new Creditcoin()),
        hypeevm: Object.freeze(new HypeEVM()),
        lineaeth: Object.freeze(new LineaETH()),
        oas: Object.freeze(new Oas()),
        og: Object.freeze(new Og()),
        ofc: Object.freeze(new Ofc()),
        okbxlayer: Object.freeze(new Xlayer()),
        opbnb: Object.freeze(new OpBNB()),
        fantom: Object.freeze(new Fantom()),
        morph: Object.freeze(new Morph()),
        morpheth: Object.freeze(new MorphETH()),
        optimism: Object.freeze(new Optimism()),
        osmo: Object.freeze(new Osmo()),
        rbtc: Object.freeze(new Rbtc()),
        rune: Object.freeze(new Rune()),
        stellar: Object.freeze(new Stellar()),
        sei: Object.freeze(new Sei()),
        seievm: Object.freeze(new SeiEvm()),
        sgb: Object.freeze(new Songbird()),
        sol: Object.freeze(new Sol()),
        sonic: Object.freeze(new Sonic()),
        sui: Object.freeze(new Sui()),
        near: Object.freeze(new Near()),
        stx: Object.freeze(new Stx()),
        somi: Object.freeze(new Somi()),
        soneium: Object.freeze(new Soneium()),
        susd: Object.freeze(new SUSD()),
        tao: Object.freeze(new Bittensor()),
        tempo: Object.freeze(new Tempo()),
        tia: Object.freeze(new Tia()),
        ton: Object.freeze(new Ton()),
        trx: Object.freeze(new Trx()),
        usdt0: Object.freeze(new Usdt0()),
        vet: Object.freeze(new Vet()),
        wemix: Object.freeze(new Wemix()),
        world: Object.freeze(new World()),
        xdc: Object.freeze(new Xdc()),
        xpl: Object.freeze(new Plasma()),
        xrp: Object.freeze(new Xrp()),
        xtz: Object.freeze(new Xtz()),
        xtzevm: Object.freeze(new XtzEvm()),
        zCash: Object.freeze(new ZCash()),
        zeta: Object.freeze(new Zeta()),
        zkSync: Object.freeze(new ZkSync()),
        zkSyncEra: Object.freeze(new ZkSyncEra()),
        unieth: Object.freeze(new Unieth()),
    },
    test: {
        ada: Object.freeze(new AdaTestnet()),
        algorand: Object.freeze(new AlgorandTestnet()),
        apechain: Object.freeze(new ApeChainTestnet()),
        apt: Object.freeze(new AptTestnet()),
        arbitrum: Object.freeze(new ArbitrumTestnet()),
        arcusdc: Object.freeze(new ArcUSDCTestnet()),
        asi: Object.freeze(new AsiTestnet()),
        atom: Object.freeze(new AtomTestnet()),
        avalancheC: Object.freeze(new AvalancheCTestnet()),
        avalancheP: Object.freeze(new AvalanchePTestnet()),
        baby: Object.freeze(new BabyTestnet()),
        basechain: Object.freeze(new BaseChainTestnet()),
        bitcoin: Object.freeze(new BitcoinTestnet()),
        bitcoinPublicSignet: Object.freeze(new BitcoinPublicSignet()),
        bitcoinTestnet4: Object.freeze(new BitcoinTestnet4()),
        bitcoinBitGoSignet: Object.freeze(new BitcoinBitGoSignet()),
        bitcoinCash: Object.freeze(new BitcoinCashTestnet()),
        bitcoinGold: Object.freeze(new BitcoinGoldTestnet()),
        bitcoinSV: Object.freeze(new BitcoinSVTestnet()),
        bera: Object.freeze(new BerachainTestnet()),
        bld: Object.freeze(new BldTestnet()),
        bsc: Object.freeze(new BinanceSmartChainTestnet()),
        canton: Object.freeze(new CantonTestnet()),
        casper: Object.freeze(new CasperTestnet()),
        celo: Object.freeze(new CeloTestnet()),
        chiliz: Object.freeze(new ChilizTestnet()),
        codexeth: Object.freeze(new CodexEthTestnet()),
        coredao: Object.freeze(new CoredaoTestnet()),
        cronos: Object.freeze(new CronosTestnet()),
        dash: Object.freeze(new DashTestnet()),
        dogecoin: Object.freeze(new DogecoinTestnet()),
        dogeos: Object.freeze(new DogeosTestnet()),
        dot: Object.freeze(new WestendAssetHub()),
        dydx: Object.freeze(new DydxTestnet()),
        eCash: Object.freeze(new ECashTestnet()),
        eos: Object.freeze(new EosTestnet()),
        fiat: Object.freeze(new FiatTestnet()),
        fetchai: Object.freeze(new FetchAiTestnet()),
        flow: Object.freeze(new FlowTestnet()),
        flr: Object.freeze(new FlareTestnet()),
        flrP: Object.freeze(new FlarePTestnet()),
        fluenteth: Object.freeze(new FluentETHTestnet()),
        h: Object.freeze(new HumanityProtocolTestnet()),
        mon: Object.freeze(new MonadTestnet()),
        megaeth: Object.freeze(new MegaETHTestnet()),
        pyrmont: Object.freeze(new Pyrmont()),
        plume: Object.freeze(new PlumeTestnet()),
        ethereumClassicTestnet: Object.freeze(new EthereumClassicTestnet()),
        hash: Object.freeze(new HashTestnet()),
        hedera: Object.freeze(new HederaTestnet()),
        hoodeth: Object.freeze(new HoodethTestnet()),
        hppeth: Object.freeze(new HppethTestnet()),
        hederaEVM: Object.freeze(new HederaEVMTestnet()),
        hemieth: Object.freeze(new HemiEthTestnet()),
        icp: Object.freeze(new IcpTestnet()),
        ip: Object.freeze(new IPTestnet()),
        initia: Object.freeze(new InitiaTestnet()),
        injective: Object.freeze(new InjectiveTestnet()),
        iota: Object.freeze(new IotaTestnet()),
        irys: Object.freeze(new IrysTestnet()),
        islm: Object.freeze(new IslmTestnet()),
        jovayeth: Object.freeze(new JovayETHTestnet()),
        kavacosmos: Object.freeze(new KavaCosmosTestnet()),
        kavaevm: Object.freeze(new KavaEVMTestnet()),
        kovan: Object.freeze(new Kovan()),
        goerli: Object.freeze(new Goerli()),
        holesky: Object.freeze(new Holesky()),
        hoodi: Object.freeze(new Hoodi()),
        lnbtc: Object.freeze(new LightningBitcoinTestnet()),
        litecoin: Object.freeze(new LitecoinTestnet()),
        mantle: Object.freeze(new MantleTestnet()),
        mantra: Object.freeze(new MantraTestnet()),
        polygon: Object.freeze(new PolygonTestnet()),
        polyx: Object.freeze(new PolymeshTestnet()),
        phrs: Object.freeze(new PharosTestnet()),
        ctc: Object.freeze(new CreditcoinTestnet()),
        hypeevm: Object.freeze(new HypeEVMTestnet()),
        lineaeth: Object.freeze(new LineaETHTestnet()),
        oas: Object.freeze(new OasTestnet()),
        og: Object.freeze(new OgTestnet()),
        ofc: Object.freeze(new OfcTestnet()),
        okbxlayer: Object.freeze(new XlayerTestnet()),
        opbnb: Object.freeze(new OpBNBTestnet()),
        fantom: Object.freeze(new FantomTestnet()),
        morph: Object.freeze(new MorphTestnet()),
        morpheth: Object.freeze(new MorphETHTestnet()),
        optimism: Object.freeze(new OptimismTestnet()),
        osmo: Object.freeze(new OsmoTestnet()),
        rbtc: Object.freeze(new RbtcTestnet()),
        rune: Object.freeze(new RuneTestNet()),
        sgb: Object.freeze(new SongbirdTestnet()),
        stellar: Object.freeze(new StellarTestnet()),
        sei: Object.freeze(new SeiTestnet()),
        seievm: Object.freeze(new SeiEvmTestnet()),
        sol: Object.freeze(new SolTestnet()),
        sui: Object.freeze(new SuiTestnet()),
        near: Object.freeze(new NearTestnet()),
        stx: Object.freeze(new StxTestnet()),
        stt: Object.freeze(new SomniaTestnet()),
        soneium: Object.freeze(new SoneiumTestnet()),
        sonic: Object.freeze(new SonicTestnet()),
        kaia: Object.freeze(new KaiaTestnet()),
        susd: Object.freeze(new SUSDTestnet()),
        coreum: Object.freeze(new CoreumTestnet()),
        tao: Object.freeze(new BittensorTestnet()),
        tempo: Object.freeze(new TempoTestnet()),
        tia: Object.freeze(new TiaTestnet()),
        ton: Object.freeze(new TonTestnet()),
        trx: Object.freeze(new TrxTestnet()),
        usdt0: Object.freeze(new Usdt0Testnet()),
        vet: Object.freeze(new VetTestnet()),
        wemix: Object.freeze(new WemixTestnet()),
        world: Object.freeze(new WorldTestnet()),
        xdc: Object.freeze(new XdcTestnet()),
        xpl: Object.freeze(new PlasmaTestnet()),
        xrp: Object.freeze(new XrpTestnet()),
        xtz: Object.freeze(new XtzTestnet()),
        xtzevm: Object.freeze(new XtzEvmTestnet()),
        zCash: Object.freeze(new ZCashTestnet()),
        zeta: Object.freeze(new ZetaTestnet()),
        zkSync: Object.freeze(new ZkSyncTestnet()),
        zkSyncEra: Object.freeze(new ZkSyncEraTestnet()),
        unieth: Object.freeze(new UniethTestnet()),
    },
};
