// @bitgo/statics/dist/src/tokenConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testnetTokens = exports.mainnetTokens = exports.formattedAlgoTokens = exports.tokens = exports.verifyTokens = exports.getFormattedTokens = exports.getFormattedTokensByNetwork = exports.getEthLikeTokens = exports.getFormattedEthLikeTokenConfig = exports.TokenTypeEnum = exports.getFormattedAlgoTokens = exports.ethGasConfigs = exports.getFormattedErc20Tokens = void 0;
exports.getFormattedTokenConfigForCoin = getFormattedTokenConfigForCoin;
const account_1 = require("./account");
const base_1 = require("./base");
const coins_1 = require("./coins");
const networks_1 = require("./networks");
const ofc_1 = require("./ofc");
function getErc20TokenConfig(coin) {
    let baseCoin;
    switch (coin.network.name) {
        case networks_1.Networks.main.ethereum.name:
            baseCoin = 'eth';
            break;
        case networks_1.Networks.test.kovan.name:
            baseCoin = 'teth';
            break;
        case networks_1.Networks.test.goerli.name:
            baseCoin = 'gteth';
            break;
        case networks_1.Networks.test.holesky.name:
        case networks_1.Networks.test.hoodi.name:
            baseCoin = 'hteth';
            break;
        default:
            throw new Error(`Erc20 token ${coin.name} has an unsupported network`);
    }
    return {
        type: coin.name,
        coin: baseCoin,
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
// Get the list of ERC-20 tokens from statics and format it properly
const getFormattedErc20Tokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.Erc20Coin) {
        acc.push(getErc20TokenConfig(coin));
    }
    return acc;
}, []);
exports.getFormattedErc20Tokens = getFormattedErc20Tokens;
exports.ethGasConfigs = {
    minimumGasPrice: 1000000000, // minimum gas price a user can provide (1 Gwei)
    defaultGasPrice: 20000000000, // default gas price if estimation fails (20 Gwei)
    maximumGasPrice: 2500000000000, // minimum gas price a user can provide (2500 Gwei)
    defaultGasLimit: 500000, // Default gas limit we set for contract send
    defaultGasLimitTokenSend: 1000000, // Default gas limit we set for token send
    minimumGasLimit: 30000, // minimum gas limit a user can set for a send
    maximumGasLimit: 20000000, // Customers cannot set gas limits beyond this amount
    newEthLikeCoinsMinGasLimit: 400000, // minimum gas limit a user can set for a send for eth like coins like arbitrum, optimism, etc
    opethGasL1Fees: 1000000000000000, // Buffer for opeth L1 gas fees
};
function getStellarTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'xlm' : 'txlm',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
    };
}
// Get the list of Stellar tokens from statics and format it properly
const getFormattedStellarTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.StellarCoin) {
        acc.push(getStellarTokenConfig(coin));
    }
    return acc;
}, []);
function getOfcTokenConfig(coin) {
    return {
        type: coin.name,
        coin: 'ofc',
        backingCoin: coin.asset,
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        isFiat: coin.kind === base_1.CoinKind.FIAT,
    };
}
// Get the list of OFC tokens from statics and format it properly
const getFormattedOfcCoins = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof ofc_1.OfcCoin) {
        acc.push(getOfcTokenConfig(coin));
    }
    return acc;
}, []);
function getCeloTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'celo' : 'tcelo',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedCeloTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.CeloCoin) {
        acc.push(getCeloTokenConfig(coin));
    }
    return acc;
}, []);
function getBscTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'bsc' : 'tbsc',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedBscTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.BscCoin) {
        acc.push(getBscTokenConfig(coin));
    }
    return acc;
}, []);
function getEosTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'eos' : 'teos',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractName.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
        contractName: coin.contractName,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedEosTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.EosCoin) {
        acc.push(getEosTokenConfig(coin));
    }
    return acc;
}, []);
function getAvaxCTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'avaxc' : 'tavaxc',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedAvaxCTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.AvaxERC20Token) {
        acc.push(getAvaxCTokenConfig(coin));
    }
    return acc;
}, []);
function getPolygonTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'polygon' : 'tpolygon',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedPolygonTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.PolygonERC20Token ||
        ((coin instanceof account_1.Erc721Coin || coin instanceof account_1.Erc1155Coin) && coin.family === base_1.CoinFamily.POLYGON)) {
        acc.push(getPolygonTokenConfig(coin));
    }
    return acc;
}, []);
function getSoneiumTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'soneium' : 'tsoneium',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedSoneiumTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if ((coin instanceof account_1.Erc721Coin || coin instanceof account_1.Erc1155Coin) && coin.family === base_1.CoinFamily.SONEIUM) {
        acc.push(getSoneiumTokenConfig(coin));
    }
    return acc;
}, []);
function getErc721TokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'eth' : 'hteth',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedErc721Tokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.Erc721Coin && coin.family === base_1.CoinFamily.ETH) {
        acc.push(getErc721TokenConfig(coin));
    }
    return acc;
}, []);
function getArbethTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'arbeth' : 'tarbeth',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedArbethTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.ArbethERC20Token) {
        acc.push(getArbethTokenConfig(coin));
    }
    return acc;
}, []);
function getOpethTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'opeth' : 'topeth',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedOpethTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.OpethERC20Token) {
        acc.push(getOpethTokenConfig(coin));
    }
    return acc;
}, []);
function getBaseethTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'baseeth' : 'tbaseeth',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedBaseethTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.EthLikeERC20Token && (coin.name.includes('baseeth:') || coin.name.includes('tbaseeth:'))) {
        acc.push(getBaseethTokenConfig(coin));
    }
    return acc;
}, []);
function getOgTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'og' : 'tog',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedOgTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.EthLikeERC20Token && (coin.name.includes('og:') || coin.name.includes('tog:'))) {
        acc.push(getOgTokenConfig(coin));
    }
    return acc;
}, []);
function getSeievmTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'seievm' : 'tseievm',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedSeievmTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.EthLikeERC20Token && (coin.name.includes('seievm:') || coin.name.includes('tseievm:'))) {
        acc.push(getSeievmTokenConfig(coin));
    }
    return acc;
}, []);
function getMonTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'mon' : 'tmon',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedMonTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.MonERC20Token) {
        acc.push(getMonTokenConfig(coin));
    }
    return acc;
}, []);
function getFlowTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'flow' : 'tflow',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedFlowTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.EthLikeERC20Token && (coin.name.includes('flow:') || coin.name.includes('tflow:'))) {
        acc.push(getFlowTokenConfig(coin));
    }
    return acc;
}, []);
function getLineaethTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'lineaeth' : 'tlineaeth',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedLineaethTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.EthLikeERC20Token && (coin.name.includes('lineaeth:') || coin.name.includes('tlineaeth:'))) {
        acc.push(getLineaethTokenConfig(coin));
    }
    return acc;
}, []);
function getZkethTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'zketh' : 'tzketh',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedZkethTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.ZkethERC20Token) {
        acc.push(getZkethTokenConfig(coin));
    }
    return acc;
}, []);
function getBeraTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'bera' : 'tbera',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedBeraTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.BeraERC20Token) {
        acc.push(getBeraTokenConfig(coin));
    }
    return acc;
}, []);
function getCoredaoTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'coredao' : 'tcoredao',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedCoredaoTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.CoredaoERC20Token) {
        acc.push(getCoredaoTokenConfig(coin));
    }
    return acc;
}, []);
function getWorldTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'world' : 'tworld',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedWorldTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.WorldERC20Token) {
        acc.push(getWorldTokenConfig(coin));
    }
    return acc;
}, []);
function getFlrTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'flr' : 'tflr',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedFlrTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.FlrERC20Token) {
        acc.push(getFlrTokenConfig(coin));
    }
    return acc;
}, []);
function getXdcTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'xdc' : 'txdc',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedXdcTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.XdcERC20Token) {
        acc.push(getXdcTokenConfig(coin));
    }
    return acc;
}, []);
function getSolTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'sol' : 'tsol',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenAddress: coin.tokenAddress,
        decimalPlaces: coin.decimalPlaces,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedSolTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.SolCoin) {
        acc.push(getSolTokenConfig(coin));
    }
    return acc;
}, []);
function getAlgoTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'algo' : 'talgo',
        alias: coin.alias,
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedAlgoTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.AlgoCoin) {
        acc.push(getAlgoTokenConfig(coin));
    }
    return acc;
}, []);
exports.getFormattedAlgoTokens = getFormattedAlgoTokens;
function getHbarTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'hbar' : 'thbar',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        nodeAccountId: coin.nodeAccountId,
        tokenId: coin.tokenId,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedHbarTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.HederaToken) {
        acc.push(getHbarTokenConfig(coin));
    }
    return acc;
}, []);
function getAdaTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'ada' : 'tada',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        policyId: coin.policyId,
        assetName: coin.assetName,
        decimalPlaces: coin.decimalPlaces,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedAdaTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.AdaToken) {
        acc.push(getAdaTokenConfig(coin));
    }
    return acc;
}, []);
function getTrxTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'trx' : 'ttrx',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedTrxTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.TronErc20Coin) {
        acc.push(getTrxTokenConfig(coin));
    }
    return acc;
}, []);
function getXrpTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'xrp' : 'txrp',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        issuerAddress: coin.issuerAddress,
        currencyCode: coin.currencyCode,
        domain: coin.domain,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedXrpTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.XrpCoin) {
        acc.push(getXrpTokenConfig(coin));
    }
    return acc;
}, []);
function getSuiTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'sui' : 'tsui',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        packageId: coin.packageId,
        module: coin.module,
        symbol: coin.symbol,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedSuiTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.SuiCoin) {
        acc.push(getSuiTokenConfig(coin));
    }
    return acc;
}, []);
function getTaoTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'tao' : 'ttao',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        subnetId: coin.subnetId,
    };
}
const getFormattedTaoTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.TaoCoin) {
        acc.push(getTaoTokenConfig(coin));
    }
    return acc;
}, []);
function getPolyxTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'polyx' : 'tpolyx',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        ticker: coin.ticker,
        assetId: coin.assetId,
    };
}
const getFormattedPolyxTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.PolyxCoin) {
        acc.push(getPolyxTokenConfig(coin));
    }
    return acc;
}, []);
function getAptTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'apt' : 'tapt',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        assetId: coin.assetId,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedAptTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.AptCoin) {
        acc.push(getAptTokenConfig(coin));
    }
    return acc;
}, []);
function getAptNFTCollectionConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'apt' : 'tapt',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        nftCollectionId: coin.nftCollectionId,
        decimalPlaces: coin.decimalPlaces,
    };
}
function getVetNFTCollectionConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'vet' : 'tvet',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        nftCollectionId: coin.nftCollectionId,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedAptNFTCollections = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.AptNFTCollection) {
        acc.push(getAptNFTCollectionConfig(coin));
    }
    return acc;
}, []);
const getFormattedVetNFTCollections = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.VetNFTCollection) {
        acc.push(getVetNFTCollectionConfig(coin));
    }
    return acc;
}, []);
function getSip10TokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'stx' : 'tstx',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        assetId: coin.assetId,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedSip10Tokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.Sip10Token) {
        acc.push(getSip10TokenConfig(coin));
    }
    return acc;
}, []);
function getNep141TokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'near' : 'tnear',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        contractAddress: coin.contractAddress,
        storageDepositAmount: coin.storageDepositAmount,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedNep141Tokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.Nep141Token) {
        acc.push(getNep141TokenConfig(coin));
    }
    return acc;
}, []);
function getCantonTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'canton' : 'tcanton',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        decimalPlaces: coin.decimalPlaces,
        baseUrl: coin.baseUrl,
        contractAddress: coin.contractAddress,
    };
}
const getFormattedCantonTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.CantonToken) {
        acc.push(getCantonTokenConfig(coin));
    }
    return acc;
}, []);
function getVetTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'vet' : 'tvet',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        contractAddress: coin.contractAddress,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedVetTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.VetToken) {
        acc.push(getVetTokenConfig(coin));
    }
    return acc;
}, []);
const getFormattedCosmosChainTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.CosmosChainToken) {
        acc.push(getCosmosTokenConfig(coin));
    }
    return acc;
}, []);
function getCosmosTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.name.split(':')[0].toLowerCase(),
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        denom: coin.denom,
        decimalPlaces: coin.decimalPlaces,
    };
}
function getJettonTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'ton' : 'tton',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        contractAddress: coin.contractAddress,
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedJettonTokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.JettonToken) {
        acc.push(getJettonTokenConfig(coin));
    }
    return acc;
}, []);
/**
 * Get token config for TIP-20 tokens on the Tempo network.
 * TIP-20 is the native token standard for Tempo, similar to ERC-20 on Ethereum.
 * If other chains need to support this token standard, the base coin can be determined
 * dynamically
 */
function getTip20TokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.network.type === networks_1.NetworkType.MAINNET ? 'tempo' : 'ttempo',
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
/**
 * Get all formatted TIP20 tokens
 * @param customCoinMap - Coin map to search
 */
const getFormattedTip20Tokens = (customCoinMap = coins_1.coins) => customCoinMap.reduce((acc, coin) => {
    if (coin instanceof account_1.Tip20Token) {
        acc.push(getTip20TokenConfig(coin));
    }
    return acc;
}, []);
var TokenTypeEnum;
(function (TokenTypeEnum) {
    TokenTypeEnum["ERC20"] = "erc20";
    TokenTypeEnum["ERC721"] = "erc721";
})(TokenTypeEnum || (exports.TokenTypeEnum = TokenTypeEnum = {}));
function getEthLikeTokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.name.split(':')[0].toLowerCase(),
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
function getEthLikeERC721TokenConfig(coin) {
    return {
        type: coin.name,
        coin: coin.name.split(':')[0].toLowerCase(),
        network: coin.network.type === networks_1.NetworkType.MAINNET ? 'Mainnet' : 'Testnet',
        name: coin.fullName,
        tokenContractAddress: coin.contractAddress.toString().toLowerCase(),
        decimalPlaces: coin.decimalPlaces,
    };
}
const getFormattedEthLikeTokenConfig = (customCoinMap = coins_1.coins, tokenType) => {
    let className;
    if (tokenType === TokenTypeEnum.ERC20) {
        className = account_1.EthLikeERC20Token;
    }
    else if (tokenType === TokenTypeEnum.ERC721) {
        className = account_1.EthLikeERC721Token;
    }
    else {
        throw new Error(`Unsupported token type: ${tokenType}`);
    }
    return customCoinMap.reduce((acc, coin) => {
        if (coin instanceof className) {
            if (tokenType === TokenTypeEnum.ERC20) {
                acc.push(getEthLikeTokenConfig(coin));
            }
            else if (tokenType === TokenTypeEnum.ERC721) {
                acc.push(getEthLikeERC721TokenConfig(coin));
            }
        }
        return acc;
    }, []);
};
exports.getFormattedEthLikeTokenConfig = getFormattedEthLikeTokenConfig;
/* Get all tokens of a given eth like coin for a given network */
const getEthLikeTokens = (network, tokenType) => {
    let feature;
    const networkTokens = (0, exports.getFormattedEthLikeTokenConfig)(coins_1.coins, tokenType).filter((token) => token.network === network);
    if (tokenType === TokenTypeEnum.ERC20) {
        feature = base_1.CoinFeature.SUPPORTS_ERC20;
    }
    else if (tokenType === TokenTypeEnum.ERC721) {
        feature = base_1.CoinFeature.SUPPORTS_ERC721;
    }
    const ethLikeTokenMap = {};
    coins_1.coins.forEach((coin) => {
        if (coin instanceof account_1.AccountCoin && coin.features.includes(feature)) {
            const coinName = coin.family;
            const coinNameForNetwork = network === 'Testnet' ? `t${coinName}` : coinName;
            ethLikeTokenMap[coin.family] = {
                tokens: networkTokens.filter((token) => token.coin === coinNameForNetwork),
            };
        }
    });
    return ethLikeTokenMap;
};
exports.getEthLikeTokens = getEthLikeTokens;
const mergeEthLikeTokenMap = (...maps) => {
    const mergedMap = {};
    maps.forEach((map) => {
        Object.keys(map).forEach((key) => {
            if (!mergedMap[key]) {
                mergedMap[key] = { tokens: [] };
            }
            mergedMap[key].tokens.push(...map[key].tokens);
        });
    });
    return mergedMap;
};
const getFormattedTokensByNetwork = (network, coinMap) => {
    const networkType = network === 'Mainnet' ? networks_1.NetworkType.MAINNET : networks_1.NetworkType.TESTNET;
    const ethLikeTokenMap = (0, exports.getEthLikeTokens)(network, TokenTypeEnum.ERC20);
    const ethLikeErc721TokenMap = (0, exports.getEthLikeTokens)(network, TokenTypeEnum.ERC721);
    return {
        ...mergeEthLikeTokenMap(ethLikeTokenMap, ethLikeErc721TokenMap),
        eth: {
            tokens: (0, exports.getFormattedErc20Tokens)(coinMap).filter((token) => token.network === network),
            nfts: getFormattedErc721Tokens(coinMap).filter((token) => token.network === network),
        },
        xlm: {
            tokens: getFormattedStellarTokens(coinMap).filter((token) => token.network === network),
        },
        algo: {
            tokens: (0, exports.getFormattedAlgoTokens)(coinMap).filter((token) => token.network === network),
        },
        ofc: {
            tokens: getFormattedOfcCoins(coinMap).filter((token) => coinMap.get(token.type).network.type === networkType),
        },
        celo: {
            tokens: getFormattedCeloTokens(coinMap).filter((token) => token.network === network),
        },
        bsc: {
            tokens: getFormattedBscTokens(coinMap).filter((token) => token.network === network),
        },
        eos: {
            tokens: getFormattedEosTokens(coinMap).filter((token) => token.network === network),
        },
        avaxc: {
            tokens: getFormattedAvaxCTokens(coinMap).filter((token) => token.network === network),
        },
        polygon: {
            tokens: getFormattedPolygonTokens(coinMap).filter((token) => token.network === network),
        },
        soneium: {
            tokens: getFormattedSoneiumTokens(coinMap).filter((token) => token.network === network),
        },
        arbeth: {
            tokens: getFormattedArbethTokens(coinMap).filter((token) => token.network === network),
        },
        opeth: {
            tokens: getFormattedOpethTokens(coinMap).filter((token) => token.network === network),
        },
        baseeth: {
            tokens: getFormattedBaseethTokens(coinMap).filter((token) => token.network === network),
        },
        og: {
            tokens: getFormattedOgTokens(coinMap).filter((token) => token.network === network),
        },
        flow: {
            tokens: getFormattedFlowTokens(coinMap).filter((token) => token.network === network),
        },
        mon: {
            tokens: getFormattedMonTokens(coinMap).filter((token) => token.network === network),
        },
        xdc: {
            tokens: getFormattedXdcTokens(coinMap).filter((token) => token.network === network),
        },
        lineaeth: {
            tokens: getFormattedLineaethTokens(coinMap).filter((token) => token.network === network),
        },
        seievm: {
            tokens: getFormattedSeievmTokens(coinMap).filter((token) => token.network === network),
        },
        zketh: {
            tokens: getFormattedZkethTokens(coinMap).filter((token) => token.network === network),
        },
        sol: {
            tokens: getFormattedSolTokens(coinMap).filter((token) => token.network === network),
        },
        hbar: {
            tokens: getFormattedHbarTokens(coinMap).filter((token) => token.network === network),
        },
        ada: {
            tokens: getFormattedAdaTokens(coinMap).filter((token) => token.network === network),
        },
        trx: {
            tokens: getFormattedTrxTokens(coinMap).filter((token) => token.network === network),
        },
        xrp: {
            tokens: getFormattedXrpTokens(coinMap).filter((token) => token.network === network),
        },
        sui: {
            tokens: getFormattedSuiTokens(coinMap).filter((token) => token.network === network),
        },
        tao: {
            tokens: getFormattedTaoTokens(coinMap).filter((token) => token.network === network),
        },
        polyx: {
            tokens: getFormattedPolyxTokens(coinMap).filter((token) => token.network === network),
        },
        bera: {
            tokens: getFormattedBeraTokens(coinMap).filter((token) => token.network === network),
        },
        coredao: {
            tokens: getFormattedCoredaoTokens(coinMap).filter((token) => token.network === network),
        },
        world: {
            tokens: getFormattedWorldTokens(coinMap).filter((token) => token.network === network),
        },
        flr: {
            tokens: getFormattedFlrTokens(coinMap).filter((token) => token.network === network),
        },
        stx: {
            tokens: getFormattedSip10Tokens(coinMap).filter((token) => token.network === network),
        },
        near: {
            tokens: getFormattedNep141Tokens(coinMap).filter((token) => token.network === network),
        },
        cosmos: {
            tokens: getFormattedCosmosChainTokens(coinMap).filter((token) => token.network === network),
        },
        ton: {
            tokens: getFormattedJettonTokens(coinMap).filter((token) => token.network === network),
        },
        tempo: {
            tokens: getFormattedTip20Tokens(coinMap).filter((token) => token.network === network),
        },
        canton: {
            tokens: getFormattedCantonTokens(coinMap).filter((token) => token.network === network),
        },
    };
};
exports.getFormattedTokensByNetwork = getFormattedTokensByNetwork;
const getFormattedTokens = (coinMap = coins_1.coins) => {
    const formattedAptNFTCollections = getFormattedAptNFTCollections(coinMap);
    const formattedVetNFTCollections = getFormattedVetNFTCollections(coinMap);
    return {
        bitcoin: {
            ...(0, exports.getFormattedTokensByNetwork)('Mainnet', coinMap),
            apt: {
                tokens: getFormattedAptTokens(coinMap).filter((token) => token.network === 'Mainnet'),
                nftCollections: formattedAptNFTCollections.filter((nftCollection) => nftCollection.network === 'Mainnet'),
            },
            vet: {
                tokens: getFormattedVetTokens(coinMap).filter((token) => token.network === 'Mainnet'),
                nftCollections: formattedVetNFTCollections.filter((nftCollection) => nftCollection.network === 'Mainnet'),
            },
        },
        testnet: {
            ...(0, exports.getFormattedTokensByNetwork)('Testnet', coinMap),
            apt: {
                tokens: getFormattedAptTokens(coinMap).filter((token) => token.network === 'Testnet'),
                nftCollections: formattedAptNFTCollections.filter((nftCollection) => nftCollection.network === 'Testnet'),
            },
            vet: {
                tokens: getFormattedVetTokens(coinMap).filter((token) => token.network === 'Testnet'),
                nftCollections: formattedVetNFTCollections.filter((nftCollection) => nftCollection.network === 'Testnet'),
            },
        },
    };
};
exports.getFormattedTokens = getFormattedTokens;
/**
 * Verify mainnet or testnet tokens
 * @param tokens
 */
const verifyTokens = function (tokens) {
    const verifiedTokens = {};
    tokens.forEach((token) => {
        if (verifiedTokens[token.type]) {
            throw new Error('token : ' + token.type + ' duplicated.');
        }
        verifiedTokens[token.type] = true;
        if (token.tokenContractAddress &&
            token.tokenContractAddress !==
                token.tokenContractAddress.toLocaleLowerCase()) {
            throw new Error('token contract: ' + token.type + ' is not all lower case: ' + token);
        }
    });
    return verifiedTokens;
};
exports.verifyTokens = verifyTokens;
exports.tokens = (0, exports.getFormattedTokens)();
exports.formattedAlgoTokens = (0, exports.getFormattedAlgoTokens)();
const mainnetErc20Tokens = (0, exports.verifyTokens)(exports.tokens.bitcoin.eth.tokens);
const mainnetErc721Tokens = (0, exports.verifyTokens)(exports.tokens.bitcoin.eth.nfts);
const mainnetStellarTokens = (0, exports.verifyTokens)(exports.tokens.bitcoin.xlm.tokens);
exports.mainnetTokens = { ...mainnetErc20Tokens, ...mainnetErc721Tokens, ...mainnetStellarTokens };
const testnetErc20Tokens = (0, exports.verifyTokens)(exports.tokens.testnet.eth.tokens);
const testnetErc721Tokens = (0, exports.verifyTokens)(exports.tokens.testnet.eth.nfts);
const testnetStellarTokens = (0, exports.verifyTokens)(exports.tokens.testnet.xlm.tokens);
exports.testnetTokens = { ...testnetErc20Tokens, ...testnetErc721Tokens, ...testnetStellarTokens };
/**
 * Get formatted token configuration for a single coin
 * @param coin - Static Base coin instance
 * @returns The formatted token configuration for the coin, or undefined if not supported
 */
function getFormattedTokenConfigForCoin(coin) {
    if (coin instanceof account_1.Erc20Coin) {
        return getErc20TokenConfig(coin);
    }
    else if (coin instanceof account_1.StellarCoin) {
        return getStellarTokenConfig(coin);
    }
    else if (coin instanceof ofc_1.OfcCoin) {
        return getOfcTokenConfig(coin);
    }
    else if (coin instanceof account_1.CeloCoin) {
        return getCeloTokenConfig(coin);
    }
    else if (coin instanceof account_1.BscCoin) {
        return getBscTokenConfig(coin);
    }
    else if (coin instanceof account_1.EosCoin) {
        return getEosTokenConfig(coin);
    }
    else if (coin instanceof account_1.AvaxERC20Token) {
        return getAvaxCTokenConfig(coin);
    }
    else if (coin instanceof account_1.PolygonERC20Token ||
        ((coin instanceof account_1.Erc721Coin || coin instanceof account_1.Erc1155Coin) && coin.family === base_1.CoinFamily.POLYGON)) {
        return getPolygonTokenConfig(coin);
    }
    else if ((coin instanceof account_1.Erc721Coin || coin instanceof account_1.Erc1155Coin) && coin.family === base_1.CoinFamily.SONEIUM) {
        return getSoneiumTokenConfig(coin);
    }
    else if (coin instanceof account_1.Erc721Coin && coin.family === base_1.CoinFamily.ETH) {
        return getErc721TokenConfig(coin);
    }
    else if (coin instanceof account_1.ArbethERC20Token) {
        return getArbethTokenConfig(coin);
    }
    else if (coin instanceof account_1.OpethERC20Token) {
        return getOpethTokenConfig(coin);
    }
    else if (coin instanceof account_1.ZkethERC20Token) {
        return getZkethTokenConfig(coin);
    }
    else if (coin instanceof account_1.SolCoin) {
        return getSolTokenConfig(coin);
    }
    else if (coin instanceof account_1.HederaToken) {
        return getHbarTokenConfig(coin);
    }
    else if (coin instanceof account_1.AdaToken) {
        return getAdaTokenConfig(coin);
    }
    else if (coin instanceof account_1.TronErc20Coin) {
        return getTrxTokenConfig(coin);
    }
    else if (coin instanceof account_1.XrpCoin) {
        return getXrpTokenConfig(coin);
    }
    else if (coin instanceof account_1.SuiCoin) {
        return getSuiTokenConfig(coin);
    }
    else if (coin instanceof account_1.TaoCoin) {
        return getTaoTokenConfig(coin);
    }
    else if (coin instanceof account_1.PolyxCoin) {
        return getPolyxTokenConfig(coin);
    }
    else if (coin instanceof account_1.AptCoin) {
        return getAptTokenConfig(coin);
    }
    else if (coin instanceof account_1.AptNFTCollection) {
        return getAptNFTCollectionConfig(coin);
    }
    else if (coin instanceof account_1.Sip10Token) {
        return getSip10TokenConfig(coin);
    }
    else if (coin instanceof account_1.Nep141Token) {
        return getNep141TokenConfig(coin);
    }
    else if (coin instanceof account_1.CosmosChainToken) {
        return getCosmosTokenConfig(coin);
    }
    else if (coin instanceof account_1.VetToken) {
        return getVetTokenConfig(coin);
    }
    else if (coin instanceof account_1.VetNFTCollection) {
        return getVetNFTCollectionConfig(coin);
    }
    else if (coin instanceof account_1.CoredaoERC20Token) {
        return getCoredaoTokenConfig(coin);
    }
    else if (coin instanceof account_1.JettonToken) {
        return getJettonTokenConfig(coin);
    }
    else if (coin instanceof account_1.FlrERC20Token) {
        return getFlrTokenConfig(coin);
    }
    else if (coin instanceof account_1.MonERC20Token) {
        return getMonTokenConfig(coin);
    }
    else if (coin instanceof account_1.XdcERC20Token) {
        return getXdcTokenConfig(coin);
    }
    else if (coin instanceof account_1.EthLikeERC20Token) {
        return getEthLikeTokenConfig(coin);
    }
    else if (coin instanceof account_1.EthLikeERC721Token) {
        return getEthLikeERC721TokenConfig(coin);
    }
    else if (coin instanceof account_1.CantonToken) {
        return getCantonTokenConfig(coin);
    }
    else if (coin instanceof account_1.Tip20Token) {
        return getTip20TokenConfig(coin);
    }
    return undefined;
}
