// @bitgo/sdk-coin-flrp/dist/src/lib/atomicTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtomicTransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
const flarejs_1 = require("@flarenetwork/flarejs");
const utils_1 = __importDefault(require("./utils"));
class AtomicTransactionBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.recoverSigner = false;
        this.transaction = new transaction_1.Transaction(_coinConfig);
        this.transaction._fee.fee = this.fixedFee;
    }
    /** @inheritdoc */
    async buildImplementation() {
        await this.buildFlareTransaction();
        this.setTransactionType(this.transactionType);
        if (this.hasSigner()) {
            for (const keyPair of this._signer) {
                await this.transaction.sign(keyPair);
            }
        }
        return this.transaction;
    }
    /**
     * Fee is fix for AVM atomic tx.
     *
     * @returns network.txFee
     * @protected
     */
    get fixedFee() {
        return this.transaction._network.txFee;
    }
    /**
     * Set the transaction type
     *
     * @param {TransactionType} transactionType The transaction type to be set
     */
    setTransactionType(transactionType) {
        this.transaction._type = transactionType;
    }
    /**
     * The internal chain is the one set for the coin in coinConfig.network. The external chain is the other chain involved.
     * The external chain id is the source on import and the destination on export.
     *
     * @param {string} chainId - id of the external chain
     */
    externalChainId(chainId) {
        const newTargetChainId = typeof chainId === 'string' ? utils_1.default.cb58Decode(chainId) : Buffer.from(chainId);
        this.validateChainId(newTargetChainId);
        this._externalChainId = newTargetChainId;
        return this;
    }
    /**
     * Set the transaction fee
     *
     * @param {string | bigint} feeValue - the fee value
     */
    fee(feeValue) {
        const fee = typeof feeValue === 'string' ? feeValue : feeValue.toString();
        this.transaction._fee.fee = fee;
        return this;
    }
    /**
     * Set the fee state for dynamic fee calculation (P-chain transactions)
     *
     * @param {FlrpFeeState} state - the fee state from the network
     */
    feeState(state) {
        this.transaction._feeState = state;
        return this;
    }
    /**
     * Set the amount for the transaction
     *
     * @param {bigint | string} value - the amount to transfer
     */
    amount(value) {
        const valueBigInt = typeof value === 'string' ? BigInt(value) : value;
        this.validateAmount(valueBigInt);
        this.transaction._amount = valueBigInt;
        return this;
    }
    /*
     * Key naming:
     *   A = user key, B = hsm key (BitGo), C = backup key
     *
     * _fromAddresses (BitGo convention) = [ A, B, C ] at indices [0, 1, 2]
     *
     * Signing key selection (which keys from _fromAddresses to use):
     *   - non-recovery: _fromAddresses[0] + _fromAddresses[1] (user + bitgo)
     *   - recovery: _fromAddresses[1] + _fromAddresses[2] (bitgo + backup)
     *
     * sigIndices in transaction (positions in sorted UTXO address list):
     *   - UTXO addresses are sorted by hex value on-chain
     *   - sigIndices = positions of the 2 signing keys in this sorted list
     *   - Example: if user sorts to position 2 and bitgo to position 0,
     *     then sigIndices = [0, 2] (even though we picked _fromAddresses[0, 1])
     */
    /**
     * Get the 2 signing addresses for FlareJS transaction building.
     *
     * FlareJS's matchOwners() selects signers in sorted UTXO address order,
     * not based on which keys should actually sign. By passing only 2 addresses,
     * we ensure the correct signers are selected.
     *
     * This mirrors AVAXP's approach in createInputOutput() where:
     * - For non-recovery: use user (index 0) and bitgo (index 1)
     * - For recovery: use bitgo (index 1) and recovery (index 2)
     *
     * @returns Array of 2 signing address buffers
     * @protected
     */
    getSigningAddresses() {
        const firstIndex = this.recoverSigner ? 2 : 0;
        const bitgoIndex = 1;
        if (this.transaction._fromAddresses.length < Math.max(firstIndex, bitgoIndex) + 1) {
            throw new sdk_core_1.BuildTransactionError(`Insufficient fromAddresses: need at least ${Math.max(firstIndex, bitgoIndex) + 1} addresses`);
        }
        const signingAddresses = [
            Buffer.from(this.transaction._fromAddresses[firstIndex]),
            Buffer.from(this.transaction._fromAddresses[bitgoIndex]),
        ];
        const invalidAddr = signingAddresses.find((addr) => addr.length !== 20);
        if (invalidAddr) {
            throw new sdk_core_1.BuildTransactionError(`Invalid signing address length: expected 20 bytes, got ${invalidAddr.length}`);
        }
        return signingAddresses;
    }
    /**
     * Compute addressesIndex for UTXOs.
     * addressesIndex[senderIdx] = position of sender[senderIdx] in UTXO's sorted address list.
     *
     * UTXO addresses are sorted lexicographically by byte value to match on-chain storage order.
     * @param forceRecompute - If true, recompute even if addressesIndex already exists
     * @protected
     */
    computeAddressesIndex(forceRecompute = false) {
        const sender = this.transaction._fromAddresses;
        if (!sender || sender.length === 0)
            return;
        this.transaction._utxos.forEach((utxo) => {
            if (!forceRecompute && utxo.addressesIndex && utxo.addressesIndex.length > 0) {
                return;
            }
            if (utxo.addresses && utxo.addresses.length > 0) {
                const sortedAddresses = utils_1.default.sortAddressesByHex(utxo.addresses);
                utxo.addresses = sortedAddresses;
                const utxoAddresses = sortedAddresses.map((a) => utils_1.default.parseAddress(a));
                utxo.addressesIndex = sender.map((a) => utxoAddresses.findIndex((u) => Buffer.compare(Buffer.from(u), Buffer.from(a)) === 0));
            }
        });
    }
    /**
     * Validate UTXOs have consistent addresses.
     * @protected
     */
    validateUtxoAddresses() {
        this.transaction._utxos.forEach((utxo) => {
            if (!utxo) {
                throw new sdk_core_1.BuildTransactionError('Utxo is undefined');
            }
            if (utxo.addressesIndex?.includes(-1)) {
                throw new sdk_core_1.BuildTransactionError('Addresses are inconsistent: ' + utxo.txid);
            }
        });
    }
    /**
     * Create credential for a UTXO following AVAX P approach.
     * Embed user/recovery address, leave BitGo slot empty.
     * Signing order is guaranteed: user signs first (address match), BitGo signs second (empty slot).
     *
     * @param utxo - The UTXO to create credential for
     * @param threshold - Number of signatures required
     * @param sigIndices - Optional sigIndices from FlareJS (if not provided, derived from addressesIndex)
     * @protected
     */
    createCredentialForUtxo(utxo, threshold, sigIndices) {
        const sender = this.transaction._fromAddresses;
        const addressesIndex = utxo.addressesIndex ?? [];
        const firstIndex = this.recoverSigner ? 2 : 0;
        const bitgoIndex = 1;
        if (threshold === 1) {
            if (sender && sender.length > firstIndex) {
                return new flarejs_1.Credential([utils_1.default.createEmptySigWithAddress(Buffer.from(sender[firstIndex]).toString('hex'))]);
            }
            return new flarejs_1.Credential([utils_1.default.createNewSig('')]);
        }
        if (addressesIndex.length >= 2 && sender && sender.length >= threshold) {
            const effectiveSigIndices = sigIndices && sigIndices.length >= 2
                ? sigIndices
                : [addressesIndex[firstIndex], addressesIndex[bitgoIndex]].sort((a, b) => a - b);
            const emptySignatures = [];
            for (const sigIdx of effectiveSigIndices) {
                const senderIdx = addressesIndex.findIndex((utxoPos) => utxoPos === sigIdx);
                if (senderIdx === firstIndex) {
                    emptySignatures.push(utils_1.default.createEmptySigWithAddress(Buffer.from(sender[firstIndex]).toString('hex')));
                }
                else {
                    emptySignatures.push(utils_1.default.createNewSig(''));
                }
            }
            return new flarejs_1.Credential(emptySignatures);
        }
        const emptySignatures = [];
        for (let i = 0; i < threshold; i++) {
            emptySignatures.push(utils_1.default.createNewSig(''));
        }
        return new flarejs_1.Credential(emptySignatures);
    }
    /**
     * Create AddressMap for a UTXO following AVAX P approach.
     *
     * @param utxo - The UTXO to create AddressMap for
     * @param threshold - Number of signatures required
     * @param sigIndices - Optional sigIndices from FlareJS (if not provided, derived from addressesIndex)
     * @protected
     */
    createAddressMapForUtxo(utxo, threshold, sigIndices) {
        const addressMap = new flarejs_1.utils.AddressMap();
        const sender = this.transaction._fromAddresses;
        const addressesIndex = utxo.addressesIndex ?? [];
        const firstIndex = this.recoverSigner ? 2 : 0;
        const bitgoIndex = 1;
        if (threshold === 1) {
            if (sender && sender.length > firstIndex) {
                addressMap.set(new flarejs_1.Address(sender[firstIndex]), 0);
            }
            else if (sender && sender.length > 0) {
                addressMap.set(new flarejs_1.Address(sender[0]), 0);
            }
            return addressMap;
        }
        if (addressesIndex.length >= 2 && sender && sender.length >= threshold) {
            const effectiveSigIndices = sigIndices && sigIndices.length >= 2
                ? sigIndices
                : [addressesIndex[firstIndex], addressesIndex[bitgoIndex]].sort((a, b) => a - b);
            effectiveSigIndices.forEach((sigIdx, slotIdx) => {
                const senderIdx = addressesIndex.findIndex((utxoPos) => utxoPos === sigIdx);
                if (senderIdx === bitgoIndex || senderIdx === firstIndex) {
                    addressMap.set(new flarejs_1.Address(sender[senderIdx]), slotIdx);
                }
            });
            return addressMap;
        }
        if (sender && sender.length >= threshold) {
            sender.slice(0, threshold).forEach((addr, i) => {
                addressMap.set(new flarejs_1.Address(addr), i);
            });
        }
        return addressMap;
    }
}
exports.AtomicTransactionBuilder = AtomicTransactionBuilder;
