// @bitgo/sdk-coin-flrp/dist/src/lib/ExportInCTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportInCTxBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const atomicInCTransactionBuilder_1 = require("./atomicInCTransactionBuilder");
const flarejs_1 = require("@flarenetwork/flarejs");
const utils_1 = __importDefault(require("./utils"));
const iface_1 = require("./iface");
class ExportInCTxBuilder extends atomicInCTransactionBuilder_1.AtomicInCTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * UTXOs are not required for Export Tx from C-Chain (uses EVM balance instead).
     * Override to prevent usage by throwing an error.
     *
     * @throws {BuildTransactionError} always throws as UTXOs are not applicable
     */
    decodedUtxos(_decodedUtxos) {
        throw new sdk_core_1.BuildTransactionError('UTXOs are not required for Export Tx from C-Chain');
    }
    /**
     * Set the nonce of C-Chain sender address
     *
     * @param {number | string} nonce - number that can be only used once
     */
    nonce(nonce) {
        const nonceBigInt = BigInt(nonce);
        this.validateNonce(nonceBigInt);
        this._nonce = nonceBigInt;
        return this;
    }
    /**
     * Export tx target P wallet.
     *
     * @param pAddresses
     */
    to(pAddresses) {
        const pubKeys = Array.isArray(pAddresses) ? pAddresses : pAddresses.split('~');
        this.transaction._to = pubKeys.map((addr) => utils_1.default.parseAddress(addr));
        return this;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Export;
    }
    initBuilder(tx, rawBytes, parsedCredentials) {
        const baseTx = tx;
        if (!this.verifyTxType(baseTx._type)) {
            throw new sdk_core_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
        const outputs = baseTx.exportedOutputs;
        if (outputs.length !== 1) {
            throw new sdk_core_1.BuildTransactionError('Transaction can have one output');
        }
        const output = outputs[0];
        if (Buffer.from(output.assetId.toBytes()).toString('hex') !== this.transaction._assetId) {
            throw new sdk_core_1.BuildTransactionError('AssetID mismatch');
        }
        const inputs = baseTx.ins;
        if (inputs.length !== 1) {
            throw new sdk_core_1.BuildTransactionError('Transaction can have one input');
        }
        const input = inputs[0];
        const transferOutput = output.output;
        const owners = transferOutput.getOwners();
        this.transaction._to = owners;
        const inputAmount = input.amount.value();
        const outputAmount = transferOutput.amount();
        const fee = inputAmount - outputAmount;
        this.transaction._amount = outputAmount;
        this.transaction._fee.fee = fee.toString();
        this.transaction._fromAddresses = [Buffer.from(input.address.toBytes())];
        this.transaction._locktime = transferOutput.getLocktime();
        this._nonce = input.nonce.value();
        const credentials = parsedCredentials || [];
        const hasCredentials = credentials.length > 0;
        if (hasCredentials && rawBytes) {
            this.transaction._rawSignedBytes = rawBytes;
        }
        const fromAddress = new flarejs_1.Address(this.transaction._fromAddresses[0]);
        const addressMap = new flarejs_1.utils.AddressMap([
            [fromAddress, 0],
            [fromAddress, 1],
        ]);
        const addressMaps = new flarejs_1.utils.AddressMaps([addressMap]);
        const unsignedTx = new flarejs_1.UnsignedTx(baseTx, [], addressMaps, credentials.length > 0 ? credentials : [new flarejs_1.Credential([utils_1.default.createNewSig('')])]);
        this.transaction.setTransaction(unsignedTx);
        return this;
    }
    static verifyTxType(txnType) {
        return txnType === iface_1.FlareTransactionType.EvmExportTx;
    }
    verifyTxType(txnType) {
        return ExportInCTxBuilder.verifyTxType(txnType);
    }
    /**
     * Build the export in C-chain transaction
     * @protected
     */
    buildFlareTransaction() {
        if (this.transaction.hasCredentials)
            return;
        if (this.transaction._amount === undefined) {
            throw new sdk_core_1.BuildTransactionError('amount is required');
        }
        if (this.transaction._fromAddresses.length !== 1) {
            throw new sdk_core_1.BuildTransactionError('sender is one and required');
        }
        if (this.transaction._to.length === 0) {
            throw new sdk_core_1.BuildTransactionError('to is required');
        }
        if (!this.transaction._fee.fee) {
            throw new sdk_core_1.BuildTransactionError('fee rate is required');
        }
        if (this._nonce === undefined) {
            throw new sdk_core_1.BuildTransactionError('nonce is required');
        }
        if (!this.transaction._context) {
            throw new sdk_core_1.BuildTransactionError('context is required');
        }
        const fee = BigInt(this.transaction._fee.fee);
        const fromAddressBytes = this.transaction._fromAddresses[0];
        const sortedToAddresses = [...this.transaction._to].sort((a, b) => {
            const aHex = Buffer.from(a).toString('hex');
            const bHex = Buffer.from(b).toString('hex');
            return aHex.localeCompare(bHex);
        });
        const toAddresses = sortedToAddresses.map((addr) => new flarejs_1.Address(addr));
        const exportEVMOptions = {
            threshold: this.transaction._threshold,
            locktime: this.transaction._locktime,
        };
        const exportTx = flarejs_1.evm.newExportTxFromBaseFee(this.transaction._context, fee, this.transaction._amount, this.transaction._context.pBlockchainID, fromAddressBytes, toAddresses.map((addr) => Buffer.from(addr.toBytes())), BigInt(this._nonce), utils_1.default.flareIdString(this.transaction._assetId).toString(), exportEVMOptions);
        this.transaction.setTransaction(exportTx);
    }
    /**
     * Check the nonce is non-negative.
     * @param nonce
     */
    validateNonce(nonce) {
        if (nonce < BigInt(0)) {
            throw new sdk_core_1.BuildTransactionError('Nonce must be greater or equal than 0');
        }
    }
}
exports.ExportInCTxBuilder = ExportInCTxBuilder;
