// @bitgo/sdk-coin-flrp/dist/src/lib/ExportInPTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportInPTxBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const atomicTransactionBuilder_1 = require("./atomicTransactionBuilder");
const flarejs_1 = require("@flarenetwork/flarejs");
const utils_1 = __importDefault(require("./utils"));
const iface_1 = require("./iface");
class ExportInPTxBuilder extends atomicTransactionBuilder_1.AtomicTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        // For Export FROM P-chain:
        // - external chain (destination) is C-chain
        // - blockchain ID (source) is P-chain
        this._externalChainId = utils_1.default.cb58Decode(this.transaction._network.cChainBlockchainID);
        // P-chain blockchain ID (from network config - decode from cb58 to hex)
        this.transaction._blockchainID = Buffer.from(utils_1.default.cb58Decode(this.transaction._network.blockchainID)).toString('hex');
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Export;
    }
    initBuilder(tx, rawBytes, parsedCredentials) {
        const exportTx = tx;
        if (!this.verifyTxType(exportTx._type)) {
            throw new sdk_core_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
        const outputs = exportTx.outs;
        if (outputs.length !== 1) {
            throw new sdk_core_1.BuildTransactionError('Transaction can have one external output');
        }
        const output = outputs[0];
        const outputTransfer = output.output;
        const assetId = output.assetId.toBytes();
        if (Buffer.compare(Buffer.from(assetId), Buffer.from(this.transaction._assetId, 'hex')) !== 0) {
            throw new Error('The Asset ID of the output does not match the transaction');
        }
        const outputOwners = outputTransfer.outputOwners;
        this.transaction._locktime = outputOwners.locktime.value();
        this.transaction._threshold = outputOwners.threshold.value();
        this.transaction._fromAddresses = outputOwners.addrs.map((addr) => Buffer.from(addr.toBytes()));
        this._externalChainId = Buffer.from(exportTx.destination.toBytes());
        this.transaction._amount = outputTransfer.amount();
        this.transaction._utxos = this.recoverUtxos([...exportTx.baseTx.inputs], outputOwners.addrs);
        const totalInputAmount = exportTx.baseTx.inputs.reduce((sum, input) => sum + input.amount(), BigInt(0));
        const changeOutputAmount = exportTx.baseTx.outputs.reduce((sum, out) => {
            const transferOut = out.output;
            return sum + transferOut.amount();
        }, BigInt(0));
        const fee = totalInputAmount - changeOutputAmount - this.transaction._amount;
        this.transaction._fee.fee = fee.toString();
        const credentials = parsedCredentials || [];
        const hasCredentials = credentials.length > 0;
        if (rawBytes && hasCredentials) {
            this.transaction._rawSignedBytes = rawBytes;
        }
        this.computeAddressesIndex(true);
        // Use parsed credentials if available, otherwise create new ones based on sigIndices
        // The sigIndices from the parsed transaction (stored in addressesIndex) determine
        // the correct credential ordering for on-chain verification
        const txCredentials = credentials.length > 0
            ? credentials
            : this.transaction._utxos.map((utxo) => {
                const utxoThreshold = utxo.threshold || this.transaction._threshold;
                const sigIndices = utxo.addressesIndex ?? [];
                // Use sigIndices-based method if we have valid sigIndices from parsed transaction
                if (sigIndices.length >= utxoThreshold && sigIndices.every((idx) => idx >= 0)) {
                    return this.createCredentialForUtxo(utxo, utxoThreshold, sigIndices);
                }
                return this.createCredentialForUtxo(utxo, utxoThreshold);
            });
        // Create addressMaps using sigIndices from parsed transaction for consistency
        const addressMaps = this.transaction._utxos.map((utxo) => {
            const utxoThreshold = utxo.threshold || this.transaction._threshold;
            const sigIndices = utxo.addressesIndex ?? [];
            if (sigIndices.length >= utxoThreshold && sigIndices.every((idx) => idx >= 0)) {
                return this.createAddressMapForUtxo(utxo, utxoThreshold, sigIndices);
            }
            return this.createAddressMapForUtxo(utxo, utxoThreshold);
        });
        const unsignedTx = new flarejs_1.UnsignedTx(exportTx, [], new flarejs_1.utils.AddressMaps(addressMaps), txCredentials);
        this.transaction.setTransaction(unsignedTx);
        return this;
    }
    static verifyTxType(txnType) {
        return txnType === iface_1.FlareTransactionType.PvmExportTx;
    }
    verifyTxType(txnType) {
        return ExportInPTxBuilder.verifyTxType(txnType);
    }
    /**
     * Build the export transaction for P-chain
     * Following AVAX P approach for UTXO handling and signature slot assignment.
     * @protected
     */
    async buildFlareTransaction() {
        if (this.transaction.hasCredentials)
            return;
        const feeState = this.transaction._feeState;
        if (!feeState) {
            throw new sdk_core_1.BuildTransactionError('Fee state is required');
        }
        if (!this.transaction._context) {
            throw new sdk_core_1.BuildTransactionError('context is required');
        }
        if (this.transaction._amount === undefined) {
            throw new sdk_core_1.BuildTransactionError('amount is required');
        }
        if (!this.transaction._utxos || this.transaction._utxos.length === 0) {
            throw new sdk_core_1.BuildTransactionError('UTXOs are required');
        }
        this.computeAddressesIndex();
        this.validateUtxoAddresses();
        const assetIdCb58 = utils_1.default.cb58Encode(Buffer.from(this.transaction._assetId, 'hex'));
        const nativeUtxos = utils_1.default.decodedToUtxos(this.transaction._utxos, assetIdCb58);
        const totalUtxoAmount = nativeUtxos.reduce((sum, utxo) => {
            const output = utxo.output;
            return sum + output.amount();
        }, BigInt(0));
        if (totalUtxoAmount < this.transaction._amount) {
            throw new sdk_core_1.BuildTransactionError(`Insufficient UTXO balance: have ${totalUtxoAmount.toString()} nFLR, need at least ${this.transaction._amount.toString()} nFLR (plus fee)`);
        }
        const assetId = utils_1.default.flareIdString(this.transaction._assetId).toString();
        const allFromAddresses = this.transaction._fromAddresses.map((addr) => Buffer.from(addr));
        const transferableOutput = flarejs_1.TransferableOutput.fromNative(assetId, this.transaction._amount, allFromAddresses, this.transaction._locktime, this.transaction._threshold);
        const signingAddresses = this.getSigningAddresses();
        const exportTx = flarejs_1.pvm.e.newExportTx({
            feeState,
            fromAddressesBytes: signingAddresses,
            destinationChainId: this.transaction._network.cChainBlockchainID,
            outputs: [transferableOutput],
            utxos: nativeUtxos,
        }, this.transaction._context);
        const flareUnsignedTx = exportTx;
        const innerTx = flareUnsignedTx.getTx();
        const changeOutputs = innerTx.baseTx.outputs;
        let correctedExportTx = innerTx;
        if (changeOutputs.length > 0) {
            const allWalletAddresses = this.transaction._fromAddresses.map((addr) => Buffer.from(addr));
            const correctedChangeOutputs = changeOutputs.map((output) => {
                const transferOut = output.output;
                const assetIdStr = utils_1.default.flareIdString(Buffer.from(output.assetId.toBytes()).toString('hex')).toString();
                return flarejs_1.TransferableOutput.fromNative(assetIdStr, transferOut.amount(), allWalletAddresses, this.transaction._locktime, this.transaction._threshold);
            });
            correctedExportTx = this.createCorrectedExportTx(innerTx, correctedChangeOutputs);
        }
        const utxosWithIndex = correctedExportTx.baseTx.inputs.map((input) => {
            const inputTxid = utils_1.default.cb58Encode(Buffer.from(input.utxoID.txID.toBytes()));
            const inputOutputIdx = input.utxoID.outputIdx.value().toString();
            const originalUtxo = this.transaction._utxos.find((utxo) => utxo.txid === inputTxid && utxo.outputidx === inputOutputIdx);
            if (!originalUtxo) {
                throw new sdk_core_1.BuildTransactionError(`Could not find matching UTXO for input ${inputTxid}:${inputOutputIdx}`);
            }
            const transferInput = input.input;
            const actualSigIndices = transferInput.sigIndicies();
            return {
                ...originalUtxo,
                addressesIndex: originalUtxo.addressesIndex,
                addresses: originalUtxo.addresses,
                threshold: originalUtxo.threshold || this.transaction._threshold,
                actualSigIndices,
            };
        });
        this.transaction._utxos = utxosWithIndex;
        const txCredentials = utxosWithIndex.map((utxo) => this.createCredentialForUtxo(utxo, utxo.threshold, utxo.actualSigIndices));
        const addressMaps = utxosWithIndex.map((utxo) => this.createAddressMapForUtxo(utxo, utxo.threshold, utxo.actualSigIndices));
        const fixedUnsignedTx = new flarejs_1.UnsignedTx(correctedExportTx, [], new flarejs_1.utils.AddressMaps(addressMaps), txCredentials);
        this.transaction.setTransaction(fixedUnsignedTx);
    }
    /**
     * Create a corrected ExportTx with the given change outputs.
     * This is necessary because FlareJS's newExportTx doesn't support setting
     * the threshold and locktime for change outputs.
     *
     * FlareJS declares baseTx.outputs as readonly, so we use Object.defineProperty
     * to override the property with the corrected outputs. This is a workaround until
     * FlareJS adds proper support for change output thresholds.
     *
     * @param originalTx - The original ExportTx
     * @param correctedOutputs - The corrected change outputs with proper threshold
     * @returns A new ExportTx with the corrected change outputs
     */
    createCorrectedExportTx(originalTx, correctedOutputs) {
        // FlareJS declares baseTx.outputs as `public readonly outputs: readonly TransferableOutput[]`
        // We use Object.defineProperty to override the readonly property with our corrected outputs.
        // This is necessary because FlareJS's newExportTx doesn't support change output threshold/locktime.
        Object.defineProperty(originalTx.baseTx, 'outputs', {
            value: correctedOutputs,
            writable: false,
            enumerable: true,
            configurable: true,
        });
        return originalTx;
    }
    /**
     * Recover UTXOs from inputs.
     * Uses output addresses as proxy for UTXO addresses.
     *
     * @param inputs Array of TransferableInput
     * @param outputAddrs Output owner addresses to use as proxy
     * @returns Array of decoded UTXO objects
     */
    recoverUtxos(inputs, outputAddrs) {
        const proxyAddresses = outputAddrs
            ? outputAddrs.map((addr) => utils_1.default.addressToString(this.transaction._network.hrp, this.transaction._network.alias, Buffer.from(addr.toBytes())))
            : [];
        return inputs.map((input) => {
            const utxoId = input.utxoID;
            const transferInput = input.input;
            const sigIndicies = transferInput.sigIndicies();
            const utxo = {
                outputID: iface_1.SECP256K1_Transfer_Output,
                amount: input.amount().toString(),
                txid: utils_1.default.cb58Encode(Buffer.from(utxoId.txID.toBytes())),
                outputidx: utxoId.outputIdx.value().toString(),
                threshold: sigIndicies.length || this.transaction._threshold,
                addresses: proxyAddresses,
                addressesIndex: sigIndicies,
            };
            return utxo;
        });
    }
}
exports.ExportInPTxBuilder = ExportInPTxBuilder;
