// @bitgo/sdk-coin-flrp/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeeDimensions = exports.ADDRESS_SEPARATOR = exports.SECP256K1_Transfer_Output = exports.FlareTransactionType = void 0;
/**
 * Enum for Flare transaction types
 */
var FlareTransactionType;
(function (FlareTransactionType) {
    FlareTransactionType["EvmExportTx"] = "evm.ExportTx";
    FlareTransactionType["EvmImportTx"] = "evm.ImportTx";
    FlareTransactionType["PvmExportTx"] = "pvm.ExportTx";
    FlareTransactionType["PvmImportTx"] = "pvm.ImportTx";
})(FlareTransactionType || (exports.FlareTransactionType = FlareTransactionType = {}));
/**
 * TypeId value for SECP256K1 Transfer Output
 *
 * Similar to Avalanche P-Chain's SECP256K1 Transfer Output
 */
exports.SECP256K1_Transfer_Output = 7;
exports.ADDRESS_SEPARATOR = '~';
var FeeDimensions;
(function (FeeDimensions) {
    FeeDimensions[FeeDimensions["Bandwidth"] = 0] = "Bandwidth";
    FeeDimensions[FeeDimensions["DBRead"] = 1] = "DBRead";
    FeeDimensions[FeeDimensions["DBWrite"] = 2] = "DBWrite";
    FeeDimensions[FeeDimensions["Compute"] = 3] = "Compute";
})(FeeDimensions || (exports.FeeDimensions = FeeDimensions = {}));
