// @bitgo/sdk-coin-flrp/dist/src/lib/ImportInCTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportInCTxBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const atomicInCTransactionBuilder_1 = require("./atomicInCTransactionBuilder");
const flarejs_1 = require("@flarenetwork/flarejs");
const utils_1 = __importDefault(require("./utils"));
const iface_1 = require("./iface");
class ImportInCTxBuilder extends atomicInCTransactionBuilder_1.AtomicInCTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * C-chain address who is target of the import.
     * Address format is eth like
     * @param {string} cAddress
     */
    to(cAddress) {
        this.transaction._to = [utils_1.default.parseAddress(cAddress)];
        return this;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Import;
    }
    initBuilder(tx, rawBytes, parsedCredentials) {
        const baseTx = tx;
        if (!this.verifyTxType(baseTx._type)) {
            throw new sdk_core_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
        const outputs = baseTx.Outs;
        if (outputs.length !== 1) {
            throw new sdk_core_1.BuildTransactionError('Transaction can have one output');
        }
        const output = outputs[0];
        if (Buffer.from(output.assetId.toBytes()).toString('hex') !== this.transaction._assetId) {
            throw new Error('AssetID are not equals');
        }
        this.transaction._to = [Buffer.from(output.address.toBytes())];
        const inputs = baseTx.importedInputs;
        const firstInput = inputs[0];
        const inputThreshold = firstInput.sigIndicies().length || this.transaction._threshold;
        this.transaction._threshold = inputThreshold;
        this.transaction._utxos = this.recoverUtxos(inputs);
        const totalInputAmount = inputs.reduce((t, i) => t + i.amount(), BigInt(0));
        const totalOutputAmount = output.amount.value();
        const fee = totalInputAmount - totalOutputAmount;
        const credentials = parsedCredentials || [];
        const hasCredentials = credentials.length > 0;
        if (hasCredentials && rawBytes) {
            this.transaction._rawSignedBytes = rawBytes;
        }
        this.transaction._fee = {
            fee: fee.toString(),
        };
        this.computeAddressesIndex(true);
        // Create addressMaps using sigIndices from parsed transaction for consistency
        const addressMaps = this.transaction._utxos.map((utxo) => {
            const utxoThreshold = utxo.threshold || this.transaction._threshold;
            const sigIndices = utxo.addressesIndex ?? [];
            if (sigIndices.length >= utxoThreshold && sigIndices.every((idx) => idx >= 0)) {
                return this.createAddressMapForUtxo(utxo, utxoThreshold, sigIndices);
            }
            return this.createAddressMapForUtxo(utxo, utxoThreshold);
        });
        const flareAddressMaps = new flarejs_1.utils.AddressMaps(addressMaps);
        // Use parsed credentials if available, otherwise create new ones based on sigIndices
        let txCredentials;
        if (credentials.length > 0) {
            txCredentials = credentials;
        }
        else {
            txCredentials = this.transaction._utxos.map((utxo) => {
                const utxoThreshold = utxo.threshold || this.transaction._threshold;
                const sigIndices = utxo.addressesIndex ?? [];
                if (sigIndices.length >= utxoThreshold && sigIndices.every((idx) => idx >= 0)) {
                    return this.createCredentialForUtxo(utxo, utxoThreshold, sigIndices);
                }
                return this.createCredentialForUtxo(utxo, utxoThreshold);
            });
        }
        const unsignedTx = new flarejs_1.UnsignedTx(baseTx, [], flareAddressMaps, txCredentials);
        this.transaction.setTransaction(unsignedTx);
        return this;
    }
    static verifyTxType(txnType) {
        return txnType === iface_1.FlareTransactionType.EvmImportTx;
    }
    verifyTxType(txnType) {
        return ImportInCTxBuilder.verifyTxType(txnType);
    }
    /**
     * Build the import in C-chain transaction
     * Following AVAX P approach for UTXO handling and signature slot assignment.
     * @protected
     */
    buildFlareTransaction() {
        if (this.transaction.hasCredentials)
            return;
        if (this.transaction._to.length !== 1) {
            throw new sdk_core_1.BuildTransactionError('to is required');
        }
        if (!this.transaction._fee.fee) {
            throw new sdk_core_1.BuildTransactionError('fee is required');
        }
        if (!this.transaction._context) {
            throw new sdk_core_1.BuildTransactionError('context is required');
        }
        if (!this.transaction._fromAddresses || this.transaction._fromAddresses.length === 0) {
            throw new sdk_core_1.BuildTransactionError('fromAddresses are required');
        }
        if (!this.transaction._utxos || this.transaction._utxos.length === 0) {
            throw new sdk_core_1.BuildTransactionError('UTXOs are required');
        }
        if (!this.transaction._threshold) {
            throw new sdk_core_1.BuildTransactionError('threshold is required');
        }
        this.computeAddressesIndex();
        this.validateUtxoAddresses();
        const actualFeeNFlr = BigInt(this.transaction._fee.fee);
        const sourceChain = 'P';
        // Convert decoded UTXOs to native FlareJS Utxo objects
        const assetId = utils_1.default.cb58Encode(Buffer.from(this.transaction._assetId, 'hex'));
        const nativeUtxos = utils_1.default.decodedToUtxos(this.transaction._utxos, assetId);
        // Validate UTXO balance is sufficient to cover the import fee
        const totalUtxoAmount = nativeUtxos.reduce((sum, utxo) => {
            const output = utxo.output;
            return sum + output.amount();
        }, BigInt(0));
        if (totalUtxoAmount <= actualFeeNFlr) {
            throw new sdk_core_1.BuildTransactionError(`Insufficient UTXO balance: have ${totalUtxoAmount.toString()} nFLR, need more than ${actualFeeNFlr.toString()} nFLR to cover import fee`);
        }
        const signingAddresses = this.getSigningAddresses();
        const importTx = flarejs_1.evm.newImportTx(this.transaction._context, this.transaction._to[0], signingAddresses, nativeUtxos, sourceChain, actualFeeNFlr);
        const flareUnsignedTx = importTx;
        const innerTx = flareUnsignedTx.getTx();
        const utxosWithIndex = innerTx.importedInputs.map((input) => {
            const inputTxid = utils_1.default.cb58Encode(Buffer.from(input.utxoID.txID.toBytes()));
            const inputOutputIdx = input.utxoID.outputIdx.value().toString();
            const originalUtxo = this.transaction._utxos.find((utxo) => utxo.txid === inputTxid && utxo.outputidx === inputOutputIdx);
            if (!originalUtxo) {
                throw new sdk_core_1.BuildTransactionError(`Could not find matching UTXO for input ${inputTxid}:${inputOutputIdx}`);
            }
            const transferInput = input.input;
            const actualSigIndices = transferInput.sigIndicies();
            return {
                ...originalUtxo,
                addressesIndex: originalUtxo.addressesIndex,
                addresses: originalUtxo.addresses,
                threshold: originalUtxo.threshold || this.transaction._threshold,
                actualSigIndices,
            };
        });
        this.transaction._utxos = utxosWithIndex;
        const txCredentials = utxosWithIndex.map((utxo) => this.createCredentialForUtxo(utxo, utxo.threshold, utxo.actualSigIndices));
        const addressMaps = utxosWithIndex.map((utxo) => this.createAddressMapForUtxo(utxo, utxo.threshold, utxo.actualSigIndices));
        const fixedUnsignedTx = new flarejs_1.UnsignedTx(innerTx, [], new flarejs_1.utils.AddressMaps(addressMaps), txCredentials);
        this.transaction.setTransaction(fixedUnsignedTx);
    }
    /**
     * Recover UTXOs from imported inputs.
     * Uses fromAddresses as proxy for UTXO addresses since they should be the same
     * addresses controlling the multisig.
     *
     * @param importedInputs Array of transferable inputs
     * @returns Array of decoded UTXO objects
     */
    recoverUtxos(importedInputs) {
        const proxyAddresses = this.transaction._fromAddresses && this.transaction._fromAddresses.length > 0
            ? this.transaction._fromAddresses.map((addr) => utils_1.default.addressToString(this.transaction._network.hrp, 'P', Buffer.from(addr)))
            : [];
        return importedInputs.map((input) => {
            const txid = input.utxoID.toString();
            const outputidx = input.utxoID.outputIdx.toString();
            const transferInput = input.input;
            const sigIndicies = transferInput.sigIndicies();
            return {
                outputID: iface_1.SECP256K1_Transfer_Output,
                amount: input.amount().toString(),
                txid: utils_1.default.cb58Encode(Buffer.from(txid, 'hex')),
                outputidx: outputidx,
                threshold: sigIndicies.length || this.transaction._threshold,
                addresses: proxyAddresses,
                addressesIndex: sigIndicies,
            };
        });
    }
}
exports.ImportInCTxBuilder = ImportInCTxBuilder;
