// @bitgo/sdk-coin-flrp/dist/src/lib/permissionlessValidatorTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PermissionlessValidatorTxBuilder = void 0;
const flarejs_1 = require("@flarenetwork/flarejs");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class PermissionlessValidatorTxBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(coinConfig) {
        super(coinConfig);
        this.transaction._fee.fee = this.transaction._network.txFee;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.AddPermissionlessValidator;
    }
    // Validation methods
    validateLocktime(locktime) {
        if (locktime < BigInt(0)) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: locktime must be 0 or higher');
        }
    }
    validateDelegationFeeRate(delegationFeeRate) {
        if (delegationFeeRate < Number(this.transaction._network.minDelegationFee)) {
            throw new sdk_core_1.BuildTransactionError(`Delegation fee cannot be less than ${this.transaction._network.minDelegationFee}`);
        }
    }
    validateNodeID(nodeID) {
        if (!nodeID) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: missing nodeID');
        }
        if (nodeID.slice(0, 6) !== 'NodeID') {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: invalid NodeID tag');
        }
        if (!(flarejs_1.utils.base58.decode(nodeID.slice(7)).length === 24)) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: NodeID is not in cb58 format');
        }
    }
    validateStakeDuration(startTime, endTime) {
        if (endTime < startTime) {
            throw new sdk_core_1.BuildTransactionError('End date cannot be less than start date');
        }
    }
    validateStakeAmount(amount) {
        const minStake = BigInt(this.transaction._network.minStake);
        if (amount < minStake) {
            throw new sdk_core_1.BuildTransactionError('Minimum staking amount is ' + Number(minStake) / 1000000000 + ' FLR.');
        }
    }
    // Builder methods
    rewardAddresses(address) {
        const rewardAddresses = address instanceof Array ? address : [address];
        this.transaction._rewardAddresses = rewardAddresses.map(utils_1.default.parseAddress);
        return this;
    }
    nodeID(nodeID) {
        this.validateNodeID(nodeID);
        this._nodeID = nodeID;
        return this;
    }
    blsPublicKey(blsPublicKey) {
        (0, sdk_core_1.isValidBLSPublicKey)(blsPublicKey);
        this._blsPublicKey = blsPublicKey;
        return this;
    }
    blsSignature(blsSignature) {
        (0, sdk_core_1.isValidBLSSignature)(blsSignature);
        this._blsSignature = blsSignature;
        return this;
    }
    locktime(value) {
        this.validateLocktime(BigInt(value));
        this._transaction._locktime = BigInt(value);
        return this;
    }
    delegationFeeRate(value) {
        this.validateDelegationFeeRate(value);
        this._delegationFeeRate = value;
        return this;
    }
    startTime(value) {
        this._startTime = BigInt(value);
        return this;
    }
    endTime(value) {
        this._endTime = BigInt(value);
        return this;
    }
    stakeAmount(value) {
        const valueBigInt = typeof value === 'bigint' ? value : BigInt(value);
        this.validateStakeAmount(valueBigInt);
        this._stakeAmount = valueBigInt;
        return this;
    }
    async buildImplementation() {
        this.buildFlareTransaction();
        this.transaction.setTransactionType(this.transactionType);
        if (this.hasSigner()) {
            for (const keyPair of this._signer) {
                await this.transaction.sign(keyPair);
            }
        }
        return this.transaction;
    }
    /**
     * Build the permissionless validator transaction
     * @protected
     */
    buildFlareTransaction() {
        throw new Error('Method not implemented.');
    }
    static verifyTxType(type) {
        return type === flarejs_1.TypeSymbols.AddPermissionlessValidatorTx;
    }
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    set transaction(transaction) {
        this._transaction = transaction;
    }
}
exports.PermissionlessValidatorTxBuilder = PermissionlessValidatorTxBuilder;
