// @bitgo/sdk-coin-flrp/dist/src/lib/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const flarejs_1 = require("@flarenetwork/flarejs");
const sdk_core_1 = require("@bitgo/sdk-core");
const keyPair_1 = require("./keyPair");
const transaction_1 = require("./transaction");
const utils_1 = __importDefault(require("./utils"));
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.recoverSigner = false;
        this._signer = [];
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     * @returns itself
     */
    initBuilder(tx) {
        const baseTx = tx.tx.baseTx;
        // Validate network and blockchain IDs match
        if (baseTx.NetworkId.value() !== this._transaction._networkID ||
            baseTx.BlockchainId.value() !== this._transaction._blockchainID) {
            throw new Error('Network or blockchain ID mismatch');
        }
        this._transaction.setTransaction(tx);
        return this;
    }
    // region Validators
    /**
     * Validates the threshold for multi-signature transactions
     * @param threshold - Number of required signatures
     */
    validateThreshold(threshold) {
        if (!threshold || threshold !== 2) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: threshold must be set to 2');
        }
    }
    /**
     * Validates the locktime value
     * @param locktime - Timestamp after which the output can be spent
     */
    validateLocktime(locktime) {
        if (locktime < BigInt(0)) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction: locktime must be 0 or higher');
        }
    }
    // endregion
    /**
     * Sets the threshold for multi-signature transactions
     * @param value - Number of required signatures
     */
    threshold(value) {
        this.validateThreshold(value);
        this._transaction._threshold = value;
        return this;
    }
    /**
     * Sets the locktime for the transaction
     * @param value - Timestamp after which the output can be spent
     */
    locktime(value) {
        this.validateLocktime(BigInt(value));
        this._transaction._locktime = BigInt(value);
        return this;
    }
    /**
     * Enables recovery mode for the transaction
     * @param recoverSigner - Whether to use recovery signing
     */
    recoverMode(recoverSigner = true) {
        this.recoverSigner = recoverSigner;
        return this;
    }
    /**
     * Sets the sender's public key(s)
     * @param senderPubKey - Public key or array of public keys
     */
    fromPubKey(senderPubKey) {
        const pubKeys = Array.isArray(senderPubKey) ? senderPubKey : [senderPubKey];
        this._transaction._fromAddresses = pubKeys.map((addr) => utils_1.default.parseAddress(addr));
        return this;
    }
    /**
     * Validates an array of UTXOs
     * @param utxos - Array of UTXOs to validate
     * @throws {BuildTransactionError} if validation fails
     */
    validateUtxos(utxos) {
        if (!utxos || utxos.length === 0) {
            throw new sdk_core_1.BuildTransactionError('UTXOs array cannot be empty');
        }
        utxos.forEach((utxo, index) => {
            this.validateUtxo(utxo, index);
        });
    }
    /**
     * Validates a single UTXO
     * @param utxo - UTXO to validate
     * @param index - Index in the array for error messaging
     * @throws {BuildTransactionError} if validation fails
     */
    validateUtxo(utxo, index) {
        if (!utxo) {
            throw new sdk_core_1.BuildTransactionError(`UTXO at index ${index} is null or undefined`);
        }
        if (!utxo.utxoId) {
            throw new sdk_core_1.BuildTransactionError(`UTXO at index ${index} missing required field: utxoId`);
        }
        if (!utxo.assetId) {
            throw new sdk_core_1.BuildTransactionError(`UTXO at index ${index} missing required field: assetId`);
        }
        if (!utxo.output) {
            throw new sdk_core_1.BuildTransactionError(`UTXO at index ${index} missing required field: output`);
        }
    }
    /**
     * Sets the decoded UTXOs for the transaction.
     * UTXOs should be provided in decoded format (DecodedUtxoObj).
     * @param decodedUtxos - Array of decoded UTXO objects
     */
    decodedUtxos(decodedUtxos) {
        if (!decodedUtxos || decodedUtxos.length === 0) {
            throw new sdk_core_1.BuildTransactionError('UTXOs array cannot be empty');
        }
        this._transaction._utxos = decodedUtxos;
        return this;
    }
    context(context) {
        this._transaction._context = context;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        try {
            // Parse the raw transaction using Flare's PVM serialization
            const [tx] = flarejs_1.pvmSerial.AddPermissionlessValidatorTx.fromBytes(Buffer.from(rawTransaction, 'hex'), flarejs_1.avmSerial.getAVMManager().getDefaultCodec());
            this.initBuilder(tx);
            return this._transaction;
        }
        catch (e) {
            throw new sdk_core_1.BuildTransactionError(`Failed to parse raw transaction: ${e.message}`);
        }
    }
    /**
     * Check the buffer has 32 byte long.
     * @param chainID
     */
    validateChainId(chainID) {
        if (chainID.length !== 32) {
            throw new sdk_core_1.BuildTransactionError('Chain id are 32 byte size');
        }
    }
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    set transaction(transaction) {
        this._transaction = transaction;
    }
    /**
     * Check that fee is greater than 0.
     * @param {bigint} fee
     */
    validateFee(fee) {
        if (fee <= BigInt(0)) {
            throw new sdk_core_1.BuildTransactionError('Fee must be greater than 0');
        }
    }
    /** @inheritdoc */
    validateKey({ key }) {
        try {
            new keyPair_1.KeyPair({ prv: key });
        }
        catch (e) {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        // throw new NotImplementedError('validateTransaction not implemented');
    }
    /** @inheritdoc */
    validateValue(value) {
        if (value.isLessThan(0)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be less than zero');
        }
    }
    /** @inheritdoc */
    validateAddress(address, addressFormat) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address');
        }
    }
    /**
     * Check the raw transaction has a valid format in the blockchain context, throw otherwise.
     *
     * @param rawTransaction Transaction in any format
     */
    validateRawTransaction(rawTransaction) {
        utils_1.default.validateRawTransaction(rawTransaction);
    }
    /** @inheritdoc */
    signImplementation({ key }) {
        this._signer.push(new keyPair_1.KeyPair({ prv: key }));
        return this.transaction;
    }
    hasSigner() {
        return this._signer !== undefined && this._signer.length > 0;
    }
    /**
     * Check the amount is positive.
     * @param amount
     */
    validateAmount(amount) {
        if (amount <= BigInt(0)) {
            throw new sdk_core_1.BuildTransactionError('Amount must be greater than 0');
        }
    }
}
exports.TransactionBuilder = TransactionBuilder;
