// @bitgo/sdk-coin-flrp/dist/src/lib/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const flarejs_1 = require("@flarenetwork/flarejs");
const buffer_1 = require("buffer");
const crypto_1 = require("crypto");
const iface_1 = require("./iface");
const utils_1 = __importDefault(require("./utils"));
/**
 * Checks if a signature is empty (first 90 hex chars are zeros)
 * @param signature
 * @returns {boolean}
 */
function isEmptySignature(signature) {
    return !!signature && utils_1.default.removeHexPrefix(signature).startsWith(''.padStart(90, '0'));
}
/**
 * Checks if an empty signature has an embedded address (non-zero bytes after position 90)
 * @param signature Hex string of the signature
 */
function hasEmbeddedAddress(signature) {
    if (!isEmptySignature(signature))
        return false;
    const cleanSig = utils_1.default.removeHexPrefix(signature);
    if (cleanSig.length < 130)
        return false;
    const embeddedPart = cleanSig.substring(90, 130);
    return embeddedPart !== '0'.repeat(40);
}
/**
 * Generates a function to check if a signature slot matches a given address.
 * If signatures have embedded addresses, it matches by address.
 * Otherwise, it just finds empty slots.
 * @param signatures Array of signature hex strings
 */
function generateSelectorSignature(signatures) {
    // Check if any empty signature has an embedded address
    const hasEmbeddedAddresses = signatures.some((sig) => isEmptySignature(sig) && hasEmbeddedAddress(sig));
    if (hasEmbeddedAddresses) {
        // Look for address embedded in the empty signature (after position 90)
        return function (sig, address) {
            try {
                if (!isEmptySignature(sig)) {
                    return false;
                }
                const cleanSig = utils_1.default.removeHexPrefix(sig);
                const embeddedAddr = cleanSig.substring(90, 130).toLowerCase();
                return embeddedAddr === address.toLowerCase();
            }
            catch (e) {
                return false;
            }
        };
    }
    else {
        // Look for any empty slot (no embedded addresses)
        return function (sig, address) {
            return isEmptySignature(sig);
        };
    }
}
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this._threshold = 2;
        this._locktime = BigInt(0);
        this._fromAddresses = [];
        this._to = [];
        this._rewardAddresses = [];
        this._utxos = [];
        this._fee = { fee: '0' };
        this._network = coinConfig.network;
        // Decode cb58-encoded asset ID to hex for use in transaction serialization
        this._assetId = utils_1.default.cb58Decode(this._network.assetId).toString('hex');
        this._blockchainID = this._network.blockchainID;
        this._networkID = this._network.networkID;
    }
    get signature() {
        if (!this.hasCredentials) {
            return [];
        }
        return this.credentials[0].getSignatures().filter((s) => !isEmptySignature(s));
    }
    get credentials() {
        return this._flareTransaction?.credentials;
    }
    get hasCredentials() {
        return this.credentials !== undefined && this.credentials.length > 0;
    }
    /** @inheritdoc */
    canSign({ key }) {
        return true;
    }
    async sign(keyPair) {
        const prv = keyPair.getPrivateKey();
        if (!prv) {
            throw new sdk_core_1.SigningError('Missing private key');
        }
        if (!this._flareTransaction) {
            throw new sdk_core_1.InvalidTransactionError('empty transaction to sign');
        }
        if (!this.hasCredentials) {
            throw new sdk_core_1.InvalidTransactionError('empty credentials to sign');
        }
        const unsignedTx = this._flareTransaction;
        const unsignedBytes = unsignedTx.toBytes();
        const publicKey = flarejs_1.secp256k1.getPublicKey(prv);
        // Derive both EVM and P-chain addresses from the public key
        const evmAddressHex = new flarejs_1.Address(flarejs_1.secp256k1.publicKeyToEthAddress(publicKey)).toHex();
        // P-chain address derivation: ripemd160(sha256(publicKey))
        const sha256Hash = (0, crypto_1.createHash)('sha256').update(buffer_1.Buffer.from(publicKey)).digest();
        const pChainAddressBuffer = (0, crypto_1.createHash)('ripemd160').update(sha256Hash).digest();
        const pChainAddressHex = pChainAddressBuffer.toString('hex');
        const addressMap = unsignedTx.getAddresses();
        // Check for both EVM and P-chain address formats
        const hasMatchingAddress = addressMap.some((addr) => {
            const addrHex = buffer_1.Buffer.from(addr).toString('hex').toLowerCase();
            return (addrHex === utils_1.default.removeHexPrefix(evmAddressHex).toLowerCase() || addrHex === pChainAddressHex.toLowerCase());
        });
        const signature = await flarejs_1.secp256k1.sign(unsignedBytes, prv);
        let signatureSet = false;
        // Check if any credential has embedded addresses - if so, we MUST use address-based matching
        const hasEmbeddedAddressesInCredentials = unsignedTx.credentials.some((credential) => {
            const signatures = credential.getSignatures();
            return signatures.some((sig) => isEmptySignature(sig) && hasEmbeddedAddress(sig));
        });
        // Use address-based slot matching if:
        // 1. addressMap contains matching address, OR
        // 2. credentials have embedded addresses (prioritize this to ensure correct slot placement)
        if (hasMatchingAddress || hasEmbeddedAddressesInCredentials) {
            // Use address-based slot matching (like AVAX-P)
            let checkSign = undefined;
            for (const credential of unsignedTx.credentials) {
                const signatures = credential.getSignatures();
                if (checkSign === undefined) {
                    checkSign = generateSelectorSignature(signatures);
                }
                // Find the slot that matches this address
                for (let i = 0; i < signatures.length; i++) {
                    const sig = signatures[i];
                    // Try matching with P-chain address first, then EVM address
                    if (checkSign(sig, pChainAddressHex) || checkSign(sig, utils_1.default.removeHexPrefix(evmAddressHex).toLowerCase())) {
                        credential.setSignature(i, signature);
                        signatureSet = true;
                        // Clear raw signed bytes since we've modified the transaction
                        this._rawSignedBytes = undefined;
                        break; // Break inner loop, but continue to sign other credentials
                    }
                }
                // Don't break outer loop - continue signing ALL credentials that have a matching slot
            }
        }
        // Fallback: If address-based matching didn't work (e.g., ImportInC loaded from unsigned tx
        // where P-chain addresses aren't in addressMaps AND no embedded addresses), sign ALL empty
        // slots across ALL credentials. This handles multisig where each UTXO needs a credential
        // signed by the same key.
        if (!signatureSet) {
            for (const credential of unsignedTx.credentials) {
                const signatures = credential.getSignatures();
                for (let i = 0; i < signatures.length; i++) {
                    if (isEmptySignature(signatures[i])) {
                        credential.setSignature(i, signature);
                        signatureSet = true;
                        this._rawSignedBytes = undefined;
                        break; // Break inner loop, but continue to sign other credentials
                    }
                }
                // Don't break outer loop - continue signing ALL credentials with empty slots
            }
        }
        if (!signatureSet) {
            throw new sdk_core_1.SigningError('No matching signature slot found for this private key');
        }
    }
    toBroadcastFormat() {
        if (!this._flareTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        // If we have the original raw signed bytes, use them directly to preserve exact format
        if (this._rawSignedBytes) {
            return flarejs_1.utils.bufferToHex(this._rawSignedBytes);
        }
        const unsignedTx = this._flareTransaction;
        const signedTxBytes = unsignedTx.getSignedTx().toBytes();
        // Both P-chain and C-chain transactions include checksum (matching avaxp behavior)
        // avaxp P-chain: transaction.ts uses addChecksum() explicitly
        // avaxp C-chain: deprecatedTransaction.ts uses Tx.toStringHex() which internally adds checksum
        const rawTx = flarejs_1.utils.bufferToHex(utils_1.default.addChecksum(signedTxBytes));
        return rawTx;
    }
    toJson() {
        if (!this._flareTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        return {
            id: this.id,
            inputs: this.inputs,
            fromAddresses: this.fromAddresses,
            threshold: this._threshold,
            locktime: this._locktime.toString(),
            type: this.type,
            signatures: this.signature,
            outputs: this.outputs,
            changeOutputs: this.changeOutputs,
            sourceChain: this.sourceChain,
            destinationChain: this.destinationChain,
        };
    }
    /**
     * Get the source chain id or undefined if it's not a cross chain transfer.
     */
    get sourceChain() {
        const tx = this._flareTransaction.getTx();
        switch (this.type) {
            case sdk_core_1.TransactionType.Import:
                if (this.isTransactionForCChain) {
                    // C-chain Import: source is the chain we're importing FROM (P-chain)
                    const importTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(importTx.sourceChain.toBytes()));
                }
                else {
                    // P-chain Import: source is the chain we're importing FROM (C-chain)
                    const pvmImportTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(pvmImportTx.sourceChain.toBytes()));
                }
            case sdk_core_1.TransactionType.Export:
                if (this.isTransactionForCChain) {
                    // C-chain Export: source is C-chain (the blockchain ID)
                    const exportTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(exportTx.blockchainId.toBytes()));
                }
                else {
                    // P-chain Export: source is P-chain (the blockchain ID from baseTx)
                    const pvmExportTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(pvmExportTx.baseTx.BlockchainId.toBytes()));
                }
            default:
                return undefined;
        }
    }
    /**
     * Get the destination chain id or undefined if it's not a cross chain transfer.
     */
    get destinationChain() {
        const tx = this._flareTransaction.getTx();
        switch (this.type) {
            case sdk_core_1.TransactionType.Import:
                if (this.isTransactionForCChain) {
                    // C-chain Import: destination is C-chain (the blockchain ID)
                    const importTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(importTx.blockchainId.toBytes()));
                }
                else {
                    // P-chain Import: destination is P-chain (the blockchain ID from baseTx)
                    const pvmImportTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(pvmImportTx.baseTx.BlockchainId.toBytes()));
                }
            case sdk_core_1.TransactionType.Export:
                if (this.isTransactionForCChain) {
                    // C-chain Export: destination is P-chain (the destination chain)
                    const exportTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(exportTx.destinationChain.toBytes()));
                }
                else {
                    // P-chain Export: destination is C-chain (the destination chain)
                    const pvmExportTx = tx;
                    return this.blockchainIDtoAlias(buffer_1.Buffer.from(pvmExportTx.destination.toBytes()));
                }
            default:
                return undefined;
        }
    }
    /**
     * Convert a blockchainId buffer to string and return P or C alias if it matches any of those chains.
     * @param {Buffer} blockchainIDBuffer
     * @return {string} blockchainID or alias if exists.
     * @private
     */
    blockchainIDtoAlias(blockchainIDBuffer) {
        const blockchainId = utils_1.default.cb58Encode(blockchainIDBuffer);
        if (blockchainId === this._network.cChainBlockchainID) {
            return 'C';
        }
        if (blockchainId === this._network.blockchainID) {
            return 'P';
        }
        return blockchainId;
    }
    setTransaction(tx) {
        this._flareTransaction = tx;
    }
    /**
     * Get the underlying Flare transaction
     * @returns The Flare transaction object
     */
    getFlareTransaction() {
        return this._flareTransaction;
    }
    setTransactionType(transactionType) {
        if (![sdk_core_1.TransactionType.AddPermissionlessValidator, sdk_core_1.TransactionType.AddPermissionlessDelegator].includes(transactionType)) {
            throw new Error(`Transaction type ${transactionType} is not supported`);
        }
        this._type = transactionType;
    }
    get signablePayload() {
        return utils_1.default.sha256(this._flareTransaction.toBytes());
    }
    get id() {
        const unsignedTx = this._flareTransaction;
        const txBytes = unsignedTx.getSignedTx().toBytes();
        const bufferArray = utils_1.default.sha256(txBytes);
        return utils_1.default.cb58Encode(buffer_1.Buffer.from(bufferArray));
    }
    get fromAddresses() {
        return this._fromAddresses.map((a) => flarejs_1.utils.format(this._network.alias, this._network.hrp, a));
    }
    get rewardAddresses() {
        return this._rewardAddresses.map((a) => flarejs_1.utils.format(this._network.alias, this._network.hrp, a));
    }
    get fee() {
        return this._fee;
    }
    /**
     * Check if this transaction is for C-chain (EVM transactions)
     */
    get isTransactionForCChain() {
        const tx = this._flareTransaction.getTx();
        const txType = tx._type;
        return txType === iface_1.FlareTransactionType.EvmExportTx || txType === iface_1.FlareTransactionType.EvmImportTx;
    }
    get outputs() {
        const tx = this._flareTransaction.getTx();
        switch (this.type) {
            case sdk_core_1.TransactionType.Import:
                if (this.isTransactionForCChain) {
                    // C-chain Import: output is to a C-chain address
                    const importTx = tx;
                    return importTx.Outs.map((out) => ({
                        address: '0x' + buffer_1.Buffer.from(out.address.toBytes()).toString('hex'),
                        value: out.amount.value().toString(),
                    }));
                }
                else {
                    // P-chain Import: outputs are the baseTx.outputs (destination on P-chain)
                    const pvmImportTx = tx;
                    return pvmImportTx.baseTx.outputs.map(utils_1.default.mapOutputToEntry(this._network));
                }
            case sdk_core_1.TransactionType.Export:
                if (this.isTransactionForCChain) {
                    // C-chain Export: exported outputs go to P-chain
                    const exportTx = tx;
                    return exportTx.exportedOutputs.map(utils_1.default.mapOutputToEntry(this._network));
                }
                else {
                    // P-chain Export: exported outputs go to C-chain
                    const pvmExportTx = tx;
                    return pvmExportTx.outs.map(utils_1.default.mapOutputToEntry(this._network));
                }
            case sdk_core_1.TransactionType.AddPermissionlessValidator:
                return [
                    {
                        address: tx.subnetValidator.validator.nodeId.toString(),
                        value: tx.subnetValidator.validator.weight.toJSON(),
                    },
                ];
            case sdk_core_1.TransactionType.AddPermissionlessDelegator:
                return [
                    {
                        address: tx.subnetValidator.validator.nodeId.toString(),
                        value: tx.subnetValidator.validator.weight.toJSON(),
                    },
                ];
            default:
                return [];
        }
    }
    get changeOutputs() {
        const tx = this._flareTransaction.getTx();
        // C-chain transactions and Import transactions don't have change outputs
        if (this.isTransactionForCChain || this.type === sdk_core_1.TransactionType.Import) {
            return [];
        }
        switch (this.type) {
            case sdk_core_1.TransactionType.Export:
                // P-chain Export: change outputs are in baseTx.outputs
                return tx.baseTx.outputs.map(utils_1.default.mapOutputToEntry(this._network));
            case sdk_core_1.TransactionType.AddPermissionlessValidator:
                return tx.baseTx.outputs.map(utils_1.default.mapOutputToEntry(this._network));
            case sdk_core_1.TransactionType.AddPermissionlessDelegator:
                return tx.baseTx.outputs.map(utils_1.default.mapOutputToEntry(this._network));
            default:
                return [];
        }
    }
    get inputs() {
        const tx = this._flareTransaction.getTx();
        switch (this.type) {
            case sdk_core_1.TransactionType.Import:
                if (this.isTransactionForCChain) {
                    // C-chain Import: inputs come from P-chain (importedInputs)
                    const importTx = tx;
                    return importTx.importedInputs.map((input) => ({
                        id: utils_1.default.cb58Encode(buffer_1.Buffer.from(input.utxoID.txID.toBytes())) + ':' + input.utxoID.outputIdx.value(),
                        address: this.fromAddresses.sort().join(iface_1.ADDRESS_SEPARATOR),
                        value: input.amount().toString(),
                    }));
                }
                else {
                    // P-chain Import: inputs are ins (imported from C-chain)
                    const pvmImportTx = tx;
                    return pvmImportTx.ins.map((input) => ({
                        id: utils_1.default.cb58Encode(buffer_1.Buffer.from(input.utxoID.txID.toBytes())) + ':' + input.utxoID.outputIdx.value(),
                        address: this.fromAddresses.sort().join(iface_1.ADDRESS_SEPARATOR),
                        value: input.amount().toString(),
                    }));
                }
            case sdk_core_1.TransactionType.Export:
                if (this.isTransactionForCChain) {
                    // C-chain Export: inputs are from C-chain (EVM inputs)
                    const exportTx = tx;
                    return exportTx.ins.map((evmInput) => ({
                        address: '0x' + buffer_1.Buffer.from(evmInput.address.toBytes()).toString('hex'),
                        value: evmInput.amount.value().toString(),
                        nonce: Number(evmInput.nonce.value()),
                    }));
                }
                else {
                    // P-chain Export: inputs are from baseTx.inputs
                    const pvmExportTx = tx;
                    return pvmExportTx.baseTx.inputs.map((input) => ({
                        id: utils_1.default.cb58Encode(buffer_1.Buffer.from(input.utxoID.txID.toBytes())) + ':' + input.utxoID.outputIdx.value(),
                        address: this.fromAddresses.sort().join(iface_1.ADDRESS_SEPARATOR),
                        value: input.amount().toString(),
                    }));
                }
            case sdk_core_1.TransactionType.AddPermissionlessValidator: {
                const validatorTx = tx;
                if (validatorTx.baseTx?.inputs) {
                    return validatorTx.baseTx.inputs.map((input) => ({
                        id: utils_1.default.cb58Encode(buffer_1.Buffer.from(input.utxoID.txID.toBytes())) + ':' + input.utxoID.outputIdx.value(),
                        address: this.fromAddresses.sort().join(iface_1.ADDRESS_SEPARATOR),
                        value: input.amount().toString(),
                    }));
                }
                return [];
            }
            case sdk_core_1.TransactionType.AddPermissionlessDelegator: {
                const delegatorTx = tx;
                if (delegatorTx.baseTx?.inputs) {
                    return delegatorTx.baseTx.inputs.map((input) => ({
                        id: utils_1.default.cb58Encode(buffer_1.Buffer.from(input.utxoID.txID.toBytes())) + ':' + input.utxoID.outputIdx.value(),
                        address: this.fromAddresses.sort().join(iface_1.ADDRESS_SEPARATOR),
                        value: input.amount().toString(),
                    }));
                }
                return [];
            }
            default:
                return [];
        }
    }
    explainTransaction() {
        const txJson = this.toJson();
        const displayOrder = ['id', 'inputs', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'type'];
        const outputAmount = txJson.outputs.reduce((p, n) => p + BigInt(n.value), BigInt(0)).toString();
        const changeAmount = txJson.changeOutputs.reduce((p, n) => p + BigInt(n.value), BigInt(0)).toString();
        let rewardAddresses;
        if ([sdk_core_1.TransactionType.AddPermissionlessValidator, sdk_core_1.TransactionType.AddPermissionlessDelegator].includes(txJson.type)) {
            rewardAddresses = this.rewardAddresses;
            displayOrder.splice(6, 0, 'rewardAddresses');
        }
        return {
            displayOrder,
            id: txJson.id,
            inputs: txJson.inputs,
            outputs: txJson.outputs.map((o) => ({ address: o.address, amount: o.value })),
            outputAmount,
            changeOutputs: txJson.changeOutputs.map((o) => ({ address: o.address, amount: o.value })),
            changeAmount,
            rewardAddresses,
            fee: this.fee,
            type: txJson.type,
        };
    }
}
exports.Transaction = Transaction;
