// @bitgo/sdk-coin-flrp/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const flarejs_1 = require("@flarenetwork/flarejs");
const sdk_core_1 = require("@bitgo/sdk-core");
const ExportInPTxBuilder_1 = require("./ExportInPTxBuilder");
const ImportInPTxBuilder_1 = require("./ImportInPTxBuilder");
const ExportInCTxBuilder_1 = require("./ExportInCTxBuilder");
const ImportInCTxBuilder_1 = require("./ImportInCTxBuilder");
const permissionlessDelegatorTxBuilder_1 = require("./permissionlessDelegatorTxBuilder");
const utils_1 = __importDefault(require("./utils"));
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * Extract credentials from remaining bytes after transaction using FlareJS codec.
     * This is the proper way to parse credentials - using the codec's UnpackPrefix method.
     * @param credentialBytes Remaining bytes after the transaction (starts with numCredentials)
     * @param codec The FlareJS codec to use for unpacking
     * @returns Array of parsed credentials
     */
    extractCredentialsWithCodec(credentialBytes, codec) {
        const credentials = [];
        if (credentialBytes.length < 4) {
            return credentials;
        }
        // Skip the first 4 bytes (numCredentials as Int type)
        // The codec doesn't know about this Int, so we skip it manually
        let remainingBytes = credentialBytes.slice(4);
        let moreCredentials = true;
        do {
            try {
                const [credential, rest] = codec.UnpackPrefix(remainingBytes);
                credentials.push(credential);
                remainingBytes = rest;
            }
            catch {
                moreCredentials = false;
            }
        } while (remainingBytes.length > 0 && moreCredentials);
        return credentials;
    }
    /**
     * Parse a raw transaction buffer using the specified VM manager.
     * @param rawBuffer The raw transaction buffer
     * @param vmType The VM type to use for parsing ('EVM' or 'PVM')
     * @returns Parsed transaction and credentials, or null if parsing fails
     */
    parseWithVM(rawBuffer, vmType) {
        try {
            const manager = flarejs_1.utils.getManagerForVM(vmType);
            const [codec, txBytes] = manager.getCodecFromBuffer(rawBuffer);
            const [tx, credentialBytes] = codec.UnpackPrefix(txBytes);
            const credentials = credentialBytes.length > 4 ? this.extractCredentialsWithCodec(credentialBytes, codec) : [];
            return { tx, credentials };
        }
        catch {
            return null;
        }
    }
    /**
     * Create the appropriate transaction builder based on transaction type.
     * @param tx The parsed transaction
     * @param rawBuffer The raw transaction buffer
     * @param credentials The extracted credentials
     * @param isEVM Whether this is an EVM transaction
     * @returns The appropriate transaction builder
     */
    createBuilder(tx, rawBuffer, credentials, isEVM) {
        if (isEVM) {
            if (ExportInCTxBuilder_1.ExportInCTxBuilder.verifyTxType(tx._type)) {
                const builder = this.getExportInCBuilder();
                builder.initBuilder(tx, rawBuffer, credentials);
                return builder;
            }
            if (ImportInCTxBuilder_1.ImportInCTxBuilder.verifyTxType(tx._type)) {
                const builder = this.getImportInCBuilder();
                builder.initBuilder(tx, rawBuffer, credentials);
                return builder;
            }
        }
        else {
            if (ImportInPTxBuilder_1.ImportInPTxBuilder.verifyTxType(tx._type)) {
                const builder = this.getImportInPBuilder();
                builder.initBuilder(tx, rawBuffer, credentials);
                return builder;
            }
            if (ExportInPTxBuilder_1.ExportInPTxBuilder.verifyTxType(tx._type)) {
                const builder = this.getExportInPBuilder();
                builder.initBuilder(tx, rawBuffer, credentials);
                return builder;
            }
            if (permissionlessDelegatorTxBuilder_1.PermissionlessDelegatorTxBuilder.verifyTxType(tx._type)) {
                const builder = this.getDelegatorBuilder();
                builder.initBuilder(tx, rawBuffer, credentials);
                return builder;
            }
        }
        throw new sdk_core_1.NotSupported('Transaction type not supported');
    }
    /** @inheritdoc */
    from(raw) {
        utils_1.default.validateRawTransaction(raw);
        const rawBuffer = Buffer.from(utils_1.default.removeHexPrefix(raw), 'hex');
        // Try EVM first, then fall back to PVM
        const evmResult = this.parseWithVM(rawBuffer, 'EVM');
        if (evmResult) {
            return this.createBuilder(evmResult.tx, rawBuffer, evmResult.credentials, true);
        }
        const pvmResult = this.parseWithVM(rawBuffer, 'PVM');
        if (pvmResult) {
            return this.createBuilder(pvmResult.tx, rawBuffer, pvmResult.credentials, false);
        }
        throw new sdk_core_1.NotSupported('Transaction type not supported');
    }
    /** @inheritdoc */
    getTransferBuilder() {
        throw new sdk_core_1.NotSupported('Transfer is not supported in P Chain');
    }
    /**
     * Export Cross chain transfer
     */
    getExportInPBuilder() {
        return new ExportInPTxBuilder_1.ExportInPTxBuilder(this._coinConfig);
    }
    /**
     * Import Cross chain transfer
     */
    getImportInPBuilder() {
        return new ImportInPTxBuilder_1.ImportInPTxBuilder(this._coinConfig);
    }
    /**
     * Import in C chain Cross chain transfer
     */
    getImportInCBuilder() {
        return new ImportInCTxBuilder_1.ImportInCTxBuilder(this._coinConfig);
    }
    /**
     * Export in C chain Cross chain transfer
     */
    getExportInCBuilder() {
        return new ExportInCTxBuilder_1.ExportInCTxBuilder(this._coinConfig);
    }
    /**
     * Get builder for AddPermissionlessDelegator transactions.
     * Used for delegating stake to existing validators.
     */
    getDelegatorBuilder() {
        return new permissionlessDelegatorTxBuilder_1.PermissionlessDelegatorTxBuilder(this._coinConfig);
    }
    /** @inheritdoc */
    getWalletInitializationBuilder() {
        throw new sdk_core_1.NotSupported('Wallet initialization is not needed');
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
