// @bitgo/sdk-coin-apt/dist/src/lib/transaction/abstractTransferTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractTransferTransaction = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const transaction_1 = require("./transaction");
/**
 * This is for transactions where one sender sends coins to recipients.
 */
class AbstractTransferTransaction extends transaction_1.Transaction {
    inputsAndOutputs() {
        const totalAmount = this._recipients.reduce((accumulator, current) => accumulator.plus(current.amount), new bignumber_js_1.default('0'));
        const inputs = [
            {
                address: this.sender,
                value: totalAmount.toString(),
                coin: this._coinConfig.name,
            },
        ];
        const outputs = this._recipients.map((recipient) => {
            return {
                address: recipient.address,
                value: recipient.amount,
                coin: this._coinConfig.name,
            };
        });
        return { inputs, outputs, externalOutputs: this._recipients };
    }
    toJson() {
        return {
            ...super.toJson(),
            recipient: this.recipient,
        };
    }
}
exports.AbstractTransferTransaction = AbstractTransferTransaction;
