// @bitgo/sdk-coin-apt/dist/src/lib/transaction/abstractDelegationPoolAmountBasedTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractDelegationPoolAmountBasedTransaction = void 0;
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const abstractDelegationPoolTransaction_1 = require("./abstractDelegationPoolTransaction");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
const constants_1 = require("../constants");
/**
 * This is a convenience class for delegation pool functions with arguments [address, amount].
 *
 * Assume this class can be deleted at any time (concrete implementations remain the same).
 * Therefore, do not store objects as this type.
 * Good: `export abstract class DelegationPoolWithdrawTransaction extends AbstractDelegationPoolAmountBasedTransaction`
 * Good: `const transaction: AbstractDelegationPoolTransaction = DelegationPoolWithdrawTransaction()`
 * Bad: `const transaction: AbstractDelegationPoolAmountBasedTransaction = DelegationPoolWithdrawTransaction()`
 */
class AbstractDelegationPoolAmountBasedTransaction extends abstractDelegationPoolTransaction_1.AbstractDelegationPoolTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this.abi = {
            typeParameters: [],
            parameters: [new ts_sdk_1.TypeTagAddress(), new ts_sdk_1.TypeTagU64()],
        };
        this._assetId = constants_1.APTOS_COIN;
    }
    parseTransactionPayload(payload) {
        if (!this.isValidPayload(payload)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid transaction payload');
        }
        const { entryFunction } = payload;
        const addressArg = entryFunction.args[0];
        const amountArg = entryFunction.args[1];
        const [{ address, amount }] = utils_1.default.parseRecipients(addressArg, amountArg);
        this.validatorAddress = address;
        this.amount = amount.toString();
    }
    getTransactionPayloadData() {
        const { validatorAddress, amount } = this;
        if (validatorAddress === undefined)
            throw new Error('validatorAddress is undefined');
        if (amount === undefined)
            throw new Error('amount is undefined');
        return {
            function: this.moveFunctionId(),
            typeArguments: [],
            functionArguments: [ts_sdk_1.AccountAddress.fromString(validatorAddress), amount],
            abi: this.abi,
        };
    }
    isValidPayload(payload) {
        return (payload instanceof ts_sdk_1.TransactionPayloadEntryFunction &&
            payload.entryFunction.args.length === 2 &&
            payload.entryFunction.type_args.length === 0);
    }
}
exports.AbstractDelegationPoolAmountBasedTransaction = AbstractDelegationPoolAmountBasedTransaction;
