// @bitgo/sdk-coin-apt/dist/src/lib/transaction/customTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTransaction = void 0;
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
const validation_1 = require("../utils/validation");
const abstractTransferTransaction_1 = require("./abstractTransferTransaction");
/**
 * Transaction class for custom Aptos transactions.
 */
class CustomTransaction extends abstractTransferTransaction_1.AbstractTransferTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this._typeArguments = [];
        this._functionArguments = [];
        this._type = sdk_core_1.TransactionType.CustomTx;
    }
    /**
     * Set the custom transaction parameters
     */
    setCustomTransactionParams(params) {
        (0, validation_1.validateModuleName)(params.moduleName);
        (0, validation_1.validateFunctionName)(params.functionName);
        this._moduleName = params.moduleName;
        this._functionName = params.functionName;
        this._typeArguments = params.typeArguments || [];
        this._functionArguments = params.functionArguments || [];
        this._entryFunctionAbi = params.abi;
    }
    /**
     * Set the entry function ABI
     */
    setEntryFunctionAbi(abi) {
        this._entryFunctionAbi = abi;
    }
    /**
     * Get the full function name
     */
    get fullFunctionName() {
        if (!this._moduleName || !this._functionName) {
            return '';
        }
        return `${this._moduleName}::${this._functionName}`;
    }
    /**
     * Parse a transaction payload to extract the custom transaction data
     * Requires ABI information for proper type-aware conversion
     *
     * @param {TransactionPayload} payload - The transaction payload to parse
     */
    parseTransactionPayload(payload) {
        if (!(payload instanceof ts_sdk_1.TransactionPayloadEntryFunction)) {
            throw new sdk_core_1.InvalidTransactionError('Expected entry function payload for custom transaction');
        }
        const entryFunction = payload.entryFunction;
        // Extract function data
        const moduleAddress = entryFunction.module_name.address.toString();
        const moduleIdentifier = entryFunction.module_name.name.identifier;
        const functionIdentifier = entryFunction.function_name.identifier;
        const moduleName = `${moduleAddress}::${moduleIdentifier}`;
        // Validate
        (0, validation_1.validateModuleName)(moduleName);
        (0, validation_1.validateFunctionName)(functionIdentifier);
        this._moduleName = moduleName;
        this._functionName = functionIdentifier;
        this._typeArguments = entryFunction.type_args.map((typeArg) => typeArg.toString());
        // Parse arguments using ABI information
        // If ABI is available, parse with type awareness; otherwise store raw args for later processing
        if (this._entryFunctionAbi?.parameters) {
            this._functionArguments = entryFunction.args.map((arg, index) => {
                const paramType = this._entryFunctionAbi?.parameters?.[index];
                if (!paramType) {
                    throw new sdk_core_1.InvalidTransactionError(`Missing ABI parameter type for argument ${index} in ${this._moduleName}::${this._functionName}. ` +
                        'ABI parameter count mismatch.');
                }
                return this.convertArgumentByABI(arg, paramType);
            });
        }
        else {
            // Store raw arguments temporarily - transaction will be re-parsed when ABI is provided
            this._functionArguments = entryFunction.args.map((arg) => arg);
        }
    }
    /**
     * Generate transaction payload data
     */
    getTransactionPayloadData() {
        const functionName = this.getValidatedFullFunctionName();
        // Arguments are pre-processed during parsing phase
        const processedArguments = this._functionArguments;
        return {
            function: functionName,
            typeArguments: this._typeArguments,
            functionArguments: processedArguments,
            abi: this._entryFunctionAbi,
        };
    }
    /**
     * Convert argument based on ABI type information
     */
    convertArgumentByABI(arg, paramType) {
        // Handle primitive values (string, number, boolean)
        if (typeof arg === 'string' || typeof arg === 'number' || typeof arg === 'boolean') {
            return this.convertPrimitiveArgument(arg, paramType);
        }
        // Handle BCS-encoded data with 'data' property
        if (arg && typeof arg === 'object' && 'data' in arg && arg.data) {
            return this.convertBcsDataArgument(arg, paramType);
        }
        // Handle nested BCS structures with 'value' property
        if (arg && typeof arg === 'object' && 'value' in arg && arg.value) {
            return this.convertNestedBcsArgument(arg, paramType);
        }
        // For anything else, return as-is
        return arg;
    }
    /**
     * Convert primitive argument values based on parameter type
     */
    convertPrimitiveArgument(arg, paramType) {
        // Address conversion for hex strings
        if (paramType instanceof ts_sdk_1.TypeTagAddress && typeof arg === 'string' && arg.startsWith('0x')) {
            return utils_1.default.tryParseAccountAddress(arg);
        }
        // Type conversions based on parameter type
        if (paramType instanceof ts_sdk_1.TypeTagBool)
            return Boolean(arg);
        // Use unified numeric type handling
        if (this.isNumericType(paramType)) {
            return this.convertPrimitiveNumericArgument(arg);
        }
        return arg;
    }
    /**
     * Convert primitive numeric arguments to string
     * - Big numbers break JavaScript (precision loss)
     * - String safer for large U64/U128/U256 values
     * - Aptos SDK converts string to correct type automatically
     */
    convertPrimitiveNumericArgument(arg) {
        // Always string - safer for big numbers
        return String(arg);
    }
    /**
     * Convert BCS data argument with 'data' property
     */
    convertBcsDataArgument(arg, paramType) {
        const bytes = Array.from(arg.data);
        const hexString = utils_1.default.bytesToHex(bytes);
        return paramType instanceof ts_sdk_1.TypeTagAddress ? utils_1.default.tryParseAccountAddress(hexString) : hexString;
    }
    /**
     * Convert nested BCS argument with 'value' property
     */
    convertNestedBcsArgument(arg, paramType) {
        // Check if inner value is a Uint8Array (common for U64 arguments)
        if (arg.value.value && arg.value.value instanceof Uint8Array) {
            const bytes = Array.from(arg.value.value);
            if (this.isNumericType(paramType)) {
                return this.convertNumericArgument(bytes, paramType);
            }
            // For non-numeric types, convert to hex string
            return utils_1.default.bytesToHex(bytes);
        }
        // Simple value wrapper - fix the bug in original implementation
        if (typeof arg.value !== 'object' || !('value' in arg.value)) {
            return this.convertArgumentByABI(arg.value, paramType);
        }
        // Double nested structure with numeric keys
        const bytes = this.extractBytesFromBcsObject(arg.value.value);
        if (bytes.length === 0)
            return arg;
        let extractedValue;
        // Convert bytes based on parameter type using unified approach
        if (this.isNumericType(paramType)) {
            extractedValue = this.convertNumericArgument(bytes, paramType);
        }
        else if (paramType instanceof ts_sdk_1.TypeTagAddress) {
            extractedValue = utils_1.default.bytesToHex(bytes);
        }
        else if (paramType instanceof ts_sdk_1.TypeTagBool) {
            extractedValue = bytes[0] === 1;
        }
        else {
            extractedValue = utils_1.default.bytesToHex(bytes);
        }
        return this.convertArgumentByABI(extractedValue, paramType);
    }
    /**
     * Extract bytes from BCS object with numeric keys
     */
    extractBytesFromBcsObject(bcsObject) {
        const keys = Object.keys(bcsObject)
            .filter((k) => !isNaN(parseInt(k, 10)))
            .sort((a, b) => parseInt(a, 10) - parseInt(b, 10));
        if (keys.length === 0)
            return [];
        return keys.map((k) => bcsObject[k]);
    }
    /**
     * Get custom transaction parameters
     */
    getCustomTransactionParams() {
        return {
            moduleName: this._moduleName || '',
            functionName: this._functionName || '',
            typeArguments: this._typeArguments,
            functionArguments: this._functionArguments,
            abi: this._entryFunctionAbi,
        };
    }
    /**
     * Override recipient getter for custom transactions
     */
    get recipient() {
        if (this._recipients.length === 0) {
            return {
                address: '',
                amount: '0',
            };
        }
        return this._recipients[0];
    }
    /**
     * Get validated full function name
     */
    getValidatedFullFunctionName() {
        if (!this._moduleName || !this._functionName) {
            throw new Error('Module name and function name must be set before building transaction');
        }
        const fullName = `${this._moduleName}::${this._functionName}`;
        // Basic validation
        const fullFunctionPattern = /^(0x[a-fA-F0-9]{1,64}|[a-zA-Z_][a-zA-Z0-9_]*)::[a-zA-Z_][a-zA-Z0-9_]*::[a-zA-Z_][a-zA-Z0-9_]*$/;
        if (!fullFunctionPattern.test(fullName)) {
            throw new Error(`Invalid full function name format: "${fullName}". Expected format: "address::module::function"`);
        }
        return fullName;
    }
    /**
     * Check if a parameter type is a numeric type
     */
    isNumericType(paramType) {
        return (paramType instanceof ts_sdk_1.TypeTagU8 ||
            paramType instanceof ts_sdk_1.TypeTagU16 ||
            paramType instanceof ts_sdk_1.TypeTagU32 ||
            paramType instanceof ts_sdk_1.TypeTagU64 ||
            paramType instanceof ts_sdk_1.TypeTagU128 ||
            paramType instanceof ts_sdk_1.TypeTagU256);
    }
    /**
     * Convert byte arrays to string numbers
     * - Small types (U8/U16/U32): 1-4 bytes, parse manually
     * - Large types (U64/U128/U256): 8+ bytes, use existing utility
     * - Both return string for consistency
     */
    convertNumericArgument(bytes, paramType) {
        if (paramType instanceof ts_sdk_1.TypeTagU8 || paramType instanceof ts_sdk_1.TypeTagU16 || paramType instanceof ts_sdk_1.TypeTagU32) {
            // Small types: parse bytes manually (1-4 bytes)
            let result = 0;
            for (let i = bytes.length - 1; i >= 0; i--) {
                result = result * 256 + bytes[i];
            }
            return result.toString();
        }
        if (paramType instanceof ts_sdk_1.TypeTagU64 || paramType instanceof ts_sdk_1.TypeTagU128 || paramType instanceof ts_sdk_1.TypeTagU256) {
            // Large types: use existing method (needs 8+ bytes)
            return utils_1.default.getAmountFromPayloadArgs(new Uint8Array(bytes));
        }
        // Unknown type: convert to hex
        return utils_1.default.bytesToHex(bytes);
    }
}
exports.CustomTransaction = CustomTransaction;
