// @bitgo/sdk-coin-apt/dist/src/lib/transaction/delegationPoolWithdrawTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegationPoolWithdrawTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../constants");
const abstractDelegationPoolAmountBasedTransaction_1 = require("./abstractDelegationPoolAmountBasedTransaction");
class DelegationPoolWithdrawTransaction extends abstractDelegationPoolAmountBasedTransaction_1.AbstractDelegationPoolAmountBasedTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this._type = sdk_core_1.TransactionType.StakingWithdraw;
    }
    moveFunctionId() {
        return constants_1.DELEGATION_POOL_WITHDRAW_FUNCTION;
    }
    inputsAndOutputs() {
        const { sender, validatorAddress, amount } = this;
        if (sender === undefined)
            throw new Error('sender is undefined');
        if (validatorAddress === undefined)
            throw new Error('validatorAddress is undefined');
        if (amount === undefined)
            throw new Error('amount is undefined');
        return {
            inputs: [
                {
                    address: validatorAddress,
                    value: amount,
                    coin: this._coinConfig.name,
                },
            ],
            outputs: [
                {
                    address: sender,
                    value: amount,
                    coin: this._coinConfig.name,
                },
            ],
            externalOutputs: [],
        };
    }
}
exports.DelegationPoolWithdrawTransaction = DelegationPoolWithdrawTransaction;
