// @bitgo/sdk-coin-apt/dist/src/lib/transaction/digitalAssetTransfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DigitalAssetTransfer = void 0;
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../constants");
const abstractTransferTransaction_1 = require("./abstractTransferTransaction");
class DigitalAssetTransfer extends abstractTransferTransaction_1.AbstractTransferTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this.transferDigitalAssetAbi = {
            typeParameters: [{ constraints: [ts_sdk_1.MoveAbility.KEY] }],
            parameters: [new ts_sdk_1.TypeTagStruct((0, ts_sdk_1.objectStructTag)(new ts_sdk_1.TypeTagGeneric(0))), new ts_sdk_1.TypeTagAddress()],
        };
        this._type = sdk_core_1.TransactionType.SendNFT;
    }
    parseTransactionPayload(payload) {
        if (!(payload instanceof ts_sdk_1.TransactionPayloadEntryFunction) ||
            payload.entryFunction.args.length !== 2 ||
            payload.entryFunction.type_args.length !== 1 ||
            constants_1.DIGITAL_ASSET_TYPE_ARGUMENT !== payload.entryFunction.type_args[0].toString()) {
            throw new sdk_core_1.InvalidTransactionError('Invalid transaction payload');
        }
        const entryFunction = payload.entryFunction;
        this._assetId = entryFunction.args[0].toString();
        this.recipients = [
            {
                address: entryFunction.args[1].toString(),
                amount: constants_1.DIGITAL_ASSET_TRANSFER_AMOUNT,
            },
        ];
    }
    getTransactionPayloadData() {
        const recipientAddress = ts_sdk_1.AccountAddress.fromString(this.recipients[0].address);
        const digitalAssetAddress = ts_sdk_1.AccountAddress.fromString(this._assetId);
        return {
            function: constants_1.DIGITAL_ASSET_TRANSFER_FUNCTION,
            typeArguments: [constants_1.DIGITAL_ASSET_TYPE_ARGUMENT],
            functionArguments: [digitalAssetAddress, recipientAddress],
            abi: this.transferDigitalAssetAbi,
        };
    }
}
exports.DigitalAssetTransfer = DigitalAssetTransfer;
