// @bitgo/sdk-coin-apt/dist/src/lib/transaction/fungibleAssetTransfer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FungibleAssetTransfer = void 0;
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../constants");
const utils_1 = __importDefault(require("../utils"));
const abstractTransferTransaction_1 = require("./abstractTransferTransaction");
class FungibleAssetTransfer extends abstractTransferTransaction_1.AbstractTransferTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this.faTransferAbi = {
            typeParameters: [{ constraints: [] }],
            parameters: [(0, ts_sdk_1.parseTypeTag)(constants_1.FUNBIGLE_ASSET_TYPE_TAG), new ts_sdk_1.TypeTagAddress(), new ts_sdk_1.TypeTagU64()],
        };
        this.batchFaTransferAbi = {
            typeParameters: [],
            parameters: [
                (0, ts_sdk_1.parseTypeTag)(constants_1.BATCH_FUNGIBLE_ASSET_TYPE_TAG),
                new ts_sdk_1.TypeTagVector(new ts_sdk_1.TypeTagAddress()),
                new ts_sdk_1.TypeTagVector(new ts_sdk_1.TypeTagU64()),
            ],
        };
        this._type = sdk_core_1.TransactionType.SendToken;
    }
    parseTransactionPayload(payload) {
        if (!this.isValidPayload(payload)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid transaction payload');
        }
        const entryFunction = payload.entryFunction;
        this._assetId = entryFunction.args[0].toString();
        const addressArg = entryFunction.args[1];
        const amountArg = entryFunction.args[2];
        this.recipients = utils_1.default.parseRecipients(addressArg, amountArg);
    }
    getTransactionPayloadData() {
        const fungibleTokenAddress = this._assetId;
        if (this.recipients.length > 1) {
            return {
                function: constants_1.FUNGIBLE_ASSET_BATCH_TRANSFER_FUNCTION,
                typeArguments: [],
                functionArguments: [
                    fungibleTokenAddress,
                    this.recipients.map((recipient) => ts_sdk_1.AccountAddress.fromString(recipient.address)),
                    this.recipients.map((recipient) => recipient.amount),
                ],
                abi: this.batchFaTransferAbi,
            };
        }
        return {
            function: constants_1.FUNGIBLE_ASSET_TRANSFER_FUNCTION,
            typeArguments: [constants_1.FUNGIBLE_ASSET_TYPE_ARGUMENT],
            functionArguments: [
                fungibleTokenAddress,
                ts_sdk_1.AccountAddress.fromString(this.recipients[0].address),
                this.recipients[0].amount,
            ],
            abi: this.faTransferAbi,
        };
    }
    isValidPayload(payload) {
        return (payload instanceof ts_sdk_1.TransactionPayloadEntryFunction &&
            payload.entryFunction.args.length === 3 &&
            (payload.entryFunction.type_args.length === 0 ||
                (payload.entryFunction.type_args.length === 1 &&
                    constants_1.FUNGIBLE_ASSET_TYPE_ARGUMENT === payload.entryFunction.type_args[0].toString())));
    }
}
exports.FungibleAssetTransfer = FungibleAssetTransfer;
