// @bitgo/sdk-coin-apt/dist/src/lib/transaction/transferTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferTransaction = void 0;
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("../constants");
const utils_1 = __importDefault(require("../utils"));
const abstractTransferTransaction_1 = require("./abstractTransferTransaction");
class TransferTransaction extends abstractTransferTransaction_1.AbstractTransferTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this.coinTransferAbi = {
            typeParameters: [{ constraints: [] }],
            parameters: [new ts_sdk_1.TypeTagAddress(), new ts_sdk_1.TypeTagU64()],
        };
        this.batchCoinTransferAbi = {
            typeParameters: [{ constraints: [] }],
            parameters: [new ts_sdk_1.TypeTagVector(new ts_sdk_1.TypeTagAddress()), new ts_sdk_1.TypeTagVector(new ts_sdk_1.TypeTagU64())],
        };
        this._type = sdk_core_1.TransactionType.Send;
        this._assetId = constants_1.APTOS_COIN;
    }
    parseTransactionPayload(payload) {
        if (!this.isValidPayload(payload)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid transaction payload');
        }
        const entryFunction = payload.entryFunction;
        this._assetId = entryFunction.type_args[0].toString();
        const addressArg = entryFunction.args[0];
        const amountArg = entryFunction.args[1];
        this.recipients = utils_1.default.parseRecipients(addressArg, amountArg);
    }
    getTransactionPayloadData() {
        if (this.recipients.length > 1) {
            return {
                function: constants_1.COIN_BATCH_TRANSFER_FUNCTION,
                typeArguments: [this.assetId],
                functionArguments: [
                    this.recipients.map((recipient) => ts_sdk_1.AccountAddress.fromString(recipient.address)),
                    this.recipients.map((recipient) => recipient.amount),
                ],
                abi: this.batchCoinTransferAbi,
            };
        }
        return {
            function: constants_1.COIN_TRANSFER_FUNCTION,
            typeArguments: [this.assetId],
            functionArguments: [ts_sdk_1.AccountAddress.fromString(this.recipients[0].address), this.recipients[0].amount],
            abi: this.coinTransferAbi,
        };
    }
    isValidPayload(payload) {
        return (payload instanceof ts_sdk_1.TransactionPayloadEntryFunction &&
            payload.entryFunction.args.length === 2 &&
            payload.entryFunction.type_args.length === 1 &&
            payload.entryFunction.type_args[0].toString().length > 0);
    }
}
exports.TransferTransaction = TransferTransaction;
