// @bitgo/sdk-coin-apt/dist/src/lib/transaction/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const logger_1 = require("@bitgo/logger");
const statics_1 = require("@bitgo/statics");
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const constants_1 = require("../constants");
const utils_1 = __importDefault(require("../utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const assert_1 = __importDefault(require("assert"));
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this._maxGasAmount = ts_sdk_1.DEFAULT_MAX_GAS_AMOUNT;
        this._gasUnitPrice = constants_1.DEFAULT_GAS_UNIT_PRICE;
        this._gasUsed = 0;
        this._expirationTime = utils_1.default.getTxnExpirationTimestamp();
        this._sequenceNumber = 0;
        this._sender = ts_sdk_1.AccountAddress.ZERO.toString();
        this._recipients = [];
        this._assetId = ts_sdk_1.AccountAddress.ZERO.toString();
        this._isSimulateTxn = false;
        this._senderSignature = {
            publicKey: {
                pub: ts_sdk_1.Hex.fromHexInput(Transaction.EMPTY_PUBLIC_KEY).toString(),
            },
            signature: Transaction.EMPTY_SIGNATURE,
        };
        this._feePayerAddress = ts_sdk_1.AccountAddress.ZERO.toString();
        this._feePayerSignature = {
            publicKey: {
                pub: ts_sdk_1.Hex.fromHexInput(Transaction.EMPTY_PUBLIC_KEY).toString(),
            },
            signature: Transaction.EMPTY_SIGNATURE,
        };
    }
    /** @inheritDoc **/
    get id() {
        this.generateTxnId();
        return this._id ?? constants_1.UNAVAILABLE_TEXT;
    }
    get sender() {
        return this._sender;
    }
    set sender(value) {
        this._sender = value;
    }
    /**
     * @deprecated - use `recipients()`.
     */
    get recipient() {
        (0, assert_1.default)(this._recipients.length > 0, 'No recipients available');
        return this._recipients[0];
    }
    /**
     * @deprecated - use `recipients()`.
     */
    set recipient(value) {
        this.recipients = [value];
    }
    get recipients() {
        return this._recipients;
    }
    set recipients(value) {
        this._recipients = value;
    }
    get sequenceNumber() {
        return this._sequenceNumber;
    }
    set sequenceNumber(value) {
        this._sequenceNumber = value;
    }
    get maxGasAmount() {
        return this._maxGasAmount;
    }
    set maxGasAmount(value) {
        this._maxGasAmount = value;
    }
    get gasUnitPrice() {
        return this._gasUnitPrice;
    }
    set gasUnitPrice(value) {
        this._gasUnitPrice = value;
    }
    get gasUsed() {
        return this._gasUsed;
    }
    set gasUsed(value) {
        this._gasUsed = value;
    }
    get expirationTime() {
        return this._expirationTime;
    }
    set expirationTime(value) {
        this._expirationTime = value;
    }
    get feePayerAddress() {
        return this._feePayerAddress;
    }
    set transactionType(transactionType) {
        this._type = transactionType;
    }
    get assetId() {
        return this._assetId;
    }
    set assetId(value) {
        this._assetId = value;
    }
    get isSimulateTxn() {
        return this._isSimulateTxn;
    }
    set isSimulateTxn(value) {
        this._isSimulateTxn = value;
    }
    fromDeserializedSignedTransaction(signedTxn) {
        try {
            const rawTxn = signedTxn.raw_txn;
            this.parseTransactionPayload(rawTxn.payload);
            this._sender = rawTxn.sender.toString();
            this._sequenceNumber = utils_1.default.castToNumber(rawTxn.sequence_number);
            this._maxGasAmount = utils_1.default.castToNumber(rawTxn.max_gas_amount);
            this._gasUnitPrice = utils_1.default.castToNumber(rawTxn.gas_unit_price);
            this._expirationTime = utils_1.default.castToNumber(rawTxn.expiration_timestamp_secs);
            this._rawTransaction = rawTxn;
            this.loadInputsAndOutputs();
            const authenticator = signedTxn.authenticator;
            this._feePayerAddress = authenticator.fee_payer.address.toString();
            const senderAuthenticator = authenticator.sender;
            const senderSignature = Buffer.from(senderAuthenticator.signature.toUint8Array());
            this.addSenderSignature({ pub: senderAuthenticator.public_key.toString() }, senderSignature);
            const feePayerAuthenticator = authenticator.fee_payer.authenticator;
            const feePayerSignature = Buffer.from(feePayerAuthenticator.signature.toUint8Array());
            this.addFeePayerSignature({ pub: utils_1.default.stripHexPrefix(feePayerAuthenticator.public_key.toString()) }, feePayerSignature);
        }
        catch (e) {
            logger_1.logger.error('invalid signed transaction', e);
            throw new Error('invalid signed transaction');
        }
    }
    canSign(_key) {
        return false;
    }
    toBroadcastFormat() {
        if (!this._rawTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        return this.serialize();
    }
    serialize() {
        let senderAuthenticator;
        let feePayerAuthenticator;
        if (this.isSimulateTxn) {
            senderAuthenticator = new ts_sdk_1.AccountAuthenticatorNoAccountAuthenticator();
            feePayerAuthenticator = new ts_sdk_1.AccountAuthenticatorNoAccountAuthenticator();
        }
        else {
            const senderPublicKeyBuffer = utils_1.default.getBufferFromHexString(this._senderSignature.publicKey.pub);
            const senderPublicKey = new ts_sdk_1.Ed25519PublicKey(senderPublicKeyBuffer);
            const senderSignature = new ts_sdk_1.Ed25519Signature(this._senderSignature.signature);
            senderAuthenticator = new ts_sdk_1.AccountAuthenticatorEd25519(senderPublicKey, senderSignature);
            const feePayerPublicKeyBuffer = utils_1.default.getBufferFromHexString(this._feePayerSignature.publicKey.pub);
            const feePayerPublicKey = new ts_sdk_1.Ed25519PublicKey(feePayerPublicKeyBuffer);
            const feePayerSignature = new ts_sdk_1.Ed25519Signature(this._feePayerSignature.signature);
            feePayerAuthenticator = new ts_sdk_1.AccountAuthenticatorEd25519(feePayerPublicKey, feePayerSignature);
        }
        const txnAuthenticator = new ts_sdk_1.TransactionAuthenticatorFeePayer(senderAuthenticator, [], [], {
            address: ts_sdk_1.AccountAddress.fromString(this._feePayerAddress),
            authenticator: feePayerAuthenticator,
        });
        const signedTxn = new ts_sdk_1.SignedTransaction(this._rawTransaction, txnAuthenticator);
        return signedTxn.toString();
    }
    addSenderSignature(publicKey, signature) {
        this._signatures = [signature.toString('hex')];
        this._senderSignature = { publicKey, signature };
    }
    getFeePayerPubKey() {
        return this._feePayerSignature.publicKey.pub;
    }
    addFeePayerSignature(publicKey, signature) {
        this._feePayerSignature = { publicKey, signature };
    }
    addFeePayerAddress(address) {
        this._feePayerAddress = address;
    }
    async build() {
        await this.buildRawTransaction();
        this.generateTxnId();
        this.loadInputsAndOutputs();
    }
    loadInputsAndOutputs() {
        const { inputs, outputs } = this.inputsAndOutputs();
        this._inputs = inputs;
        this._outputs = outputs;
    }
    fromRawTransaction(rawTransaction) {
        let signedTxn;
        try {
            signedTxn = utils_1.default.deserializeSignedTransaction(rawTransaction);
        }
        catch (e) {
            throw new Error('invalid raw transaction');
        }
        this.fromDeserializedSignedTransaction(signedTxn);
    }
    /**
     * Deserializes a signed transaction hex string
     * @param {string} signedRawTransaction
     * @returns {SignedTransaction} the aptos signed transaction
     */
    static deserializeSignedTransaction(signedRawTransaction) {
        try {
            return utils_1.default.deserializeSignedTransaction(signedRawTransaction);
        }
        catch (e) {
            throw new Error('invalid raw transaction');
        }
    }
    toJson() {
        return {
            id: this.id,
            sender: this.sender,
            recipients: this.recipients,
            sequenceNumber: this.sequenceNumber,
            maxGasAmount: this.maxGasAmount,
            gasUnitPrice: this.gasUnitPrice,
            gasUsed: this.gasUsed,
            expirationTime: this.expirationTime,
            feePayer: this.feePayerAddress,
            assetId: this.assetId,
        };
    }
    getFee() {
        return new bignumber_js_1.default(this.gasUsed).multipliedBy(this.gasUnitPrice).toString();
    }
    get signablePayload() {
        return this.feePayerAddress ? this.getSignablePayloadWithFeePayer() : this.getSignablePayloadWithoutFeePayer();
    }
    /** @inheritDoc */
    explainTransaction() {
        const displayOrder = [
            'id',
            'outputs',
            'outputAmount',
            'changeOutputs',
            'changeAmount',
            'fee',
            'withdrawAmount',
            'sender',
            'type',
        ];
        const outputs = this.inputsAndOutputs().externalOutputs;
        const outputAmount = outputs
            .reduce((accumulator, current) => accumulator.plus(current.amount), new bignumber_js_1.default('0'))
            .toString();
        return {
            displayOrder,
            id: this.id,
            outputs,
            outputAmount,
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: this.getFee() },
            sender: this.sender,
            type: this.type,
        };
    }
    async buildRawTransaction() {
        const network = this._coinConfig.network.type === statics_1.NetworkType.MAINNET ? ts_sdk_1.Network.MAINNET : ts_sdk_1.Network.TESTNET;
        const aptos = new ts_sdk_1.Aptos(new ts_sdk_1.AptosConfig({ network }));
        const senderAddress = ts_sdk_1.AccountAddress.fromString(this._sender);
        const simpleTxn = await aptos.transaction.build.simple({
            sender: senderAddress,
            data: this.getTransactionPayloadData(),
            options: {
                maxGasAmount: this.maxGasAmount,
                gasUnitPrice: this.gasUnitPrice,
                expireTimestamp: this.expirationTime,
                accountSequenceNumber: this.sequenceNumber,
            },
        });
        this._rawTransaction = simpleTxn.rawTransaction;
    }
    getSignablePayloadWithFeePayer() {
        const feePayerRawTxn = new ts_sdk_1.FeePayerRawTransaction(this._rawTransaction, [], ts_sdk_1.AccountAddress.fromString(this._feePayerAddress));
        return Buffer.from((0, ts_sdk_1.generateSigningMessage)(feePayerRawTxn.bcsToBytes(), ts_sdk_1.RAW_TRANSACTION_WITH_DATA_SALT));
    }
    getSignablePayloadWithoutFeePayer() {
        return Buffer.from((0, ts_sdk_1.generateSigningMessage)(this._rawTransaction.bcsToBytes(), ts_sdk_1.RAW_TRANSACTION_SALT));
    }
    generateTxnId() {
        if (!this._senderSignature ||
            !this._senderSignature.publicKey ||
            !this._senderSignature.signature ||
            !this._feePayerSignature ||
            !this._feePayerSignature.publicKey ||
            !this._feePayerSignature.signature ||
            !this._feePayerAddress) {
            return;
        }
        const transaction = new ts_sdk_1.SimpleTransaction(this._rawTransaction, ts_sdk_1.AccountAddress.fromString(this._feePayerAddress));
        const senderPublicKey = new ts_sdk_1.Ed25519PublicKey(utils_1.default.getBufferFromHexString(this._senderSignature.publicKey.pub));
        const senderSignature = new ts_sdk_1.Ed25519Signature(this._senderSignature.signature);
        const senderAuthenticator = new ts_sdk_1.AccountAuthenticatorEd25519(senderPublicKey, senderSignature);
        const feePayerPublicKey = new ts_sdk_1.Ed25519PublicKey(utils_1.default.getBufferFromHexString(this._feePayerSignature.publicKey.pub));
        const feePayerSignature = new ts_sdk_1.Ed25519Signature(this._feePayerSignature.signature);
        const feePayerAuthenticator = new ts_sdk_1.AccountAuthenticatorEd25519(feePayerPublicKey, feePayerSignature);
        this._id = (0, ts_sdk_1.generateUserTransactionHash)({ transaction, senderAuthenticator, feePayerAuthenticator });
    }
}
exports.Transaction = Transaction;
Transaction.EMPTY_PUBLIC_KEY = Buffer.alloc(32);
Transaction.EMPTY_SIGNATURE = Buffer.alloc(64);
