// @bitgo/utxo-core/dist/cjs/test/descriptor/psbt/psbt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const descriptor_1 = require("../../../src/descriptor");
const testutil_1 = require("../../../src/testutil");
const descriptor_2 = require("../../../src/testutil/descriptor");
const sign_1 = require("../../../src/descriptor/psbt/sign");
function normalize(v) {
    if (typeof v === 'bigint') {
        return v.toString();
    }
    if (v instanceof wasm_utxo_1.Descriptor) {
        return v.toString();
    }
    if (v instanceof Buffer) {
        return v.toString('hex');
    }
    if (v instanceof utxolib.Psbt) {
        return (0, descriptor_2.toPlainObjectFromPsbt)(v);
    }
    if (v instanceof utxolib.Transaction) {
        return (0, descriptor_2.toPlainObjectFromTx)(v);
    }
    if (Array.isArray(v)) {
        return v.map(normalize);
    }
    if (typeof v === 'object' && v !== null) {
        return Object.fromEntries(Object.entries(v).flatMap(([k, v]) => (v === undefined ? [] : [[k, normalize(v)]])));
    }
    return v;
}
async function assertEqualsFixture(t, filename, value) {
    filename = __dirname + '/../../descriptor/psbt/fixtures/' + t + '.' + filename;
    const nv = normalize(value);
    assert_1.default.deepStrictEqual(nv, await (0, testutil_1.getFixture)(filename, nv));
}
const selfKeys = (0, testutil_1.getKeyTriple)('a');
const otherKeys = (0, testutil_1.getKeyTriple)('b');
function toPlain(k) {
    assert_1.default.ok(k.privateKey);
    return utxo_lib_1.ECPair.fromPrivateKey(k.privateKey);
}
function getStages(psbt, parsed, stages) {
    return Object.fromEntries(stages.map((stage) => {
        const psbtStageWrapped = (0, descriptor_1.toWrappedPsbt)(psbt);
        for (const key of stage.keys) {
            (0, assert_1.default)((0, sign_1.getNewSignatureCount)((0, sign_1.signWithKey)(psbtStageWrapped, key)) > 0, 'No new signatures were created');
        }
        const psbtStage = (0, descriptor_1.toUtxoPsbt)(psbtStageWrapped, utxolib.networks.bitcoin);
        let psbtFinal;
        let networkTx;
        let networkTxBuffer;
        if (stage.final) {
            psbtStageWrapped.finalize();
            psbtFinal = (0, descriptor_1.toUtxoPsbt)(psbtStageWrapped, utxolib.networks.bitcoin);
            networkTx = psbtFinal.extractTransaction();
            networkTxBuffer = networkTx.toBuffer();
        }
        return [
            stage.name,
            {
                psbt: psbtStage,
                parsed,
                psbtFinal,
                networkTx,
                networkTxBuffer,
            },
        ];
    }));
}
function describeCreatePsbt(name, testParams) {
    describe(`createPsbt ${name}`, function () {
        it('creates psbt with expected properties', async function () {
            const psbtUnsigned = 'psbt' in testParams
                ? testParams.psbt
                : (0, descriptor_2.mockPsbtDefault)({
                    descriptorSelf: testParams.descriptorSelf,
                    descriptorOther: (0, descriptor_2.getDescriptor)('Wsh2Of3', otherKeys),
                    params: testParams.psbtParams,
                });
            const descriptorMap = new Map([['self', testParams.descriptorSelf]]);
            const parsed = (0, descriptor_1.parse)(psbtUnsigned, descriptorMap, utxolib.networks.bitcoin);
            assert_1.default.strictEqual(parsed.spendAmount, psbtUnsigned.txOutputs[1].value);
            await assertEqualsFixture(name, 'psbtStages.json', getStages(psbtUnsigned, parsed, testParams.stages));
        });
    });
}
const defaultStagesCombinedAB = [
    { name: 'unsigned', keys: [] },
    { name: 'signedA', keys: selfKeys.slice(0, 1) },
    { name: 'signedAB', keys: selfKeys.slice(0, 2), final: true },
];
function getDefaultStagesSeparateAB({ plain = false } = {}) {
    const keys = plain ? selfKeys.map(toPlain) : selfKeys;
    return [
        { name: 'unsigned', keys: [] },
        { name: 'signedA', keys: keys.slice(0, 1) },
        { name: 'signedB', keys: keys.slice(1, 2), final: true },
    ];
}
function describeCreatePsbt2Of3(t) {
    describeCreatePsbt(t, {
        descriptorSelf: (0, descriptor_2.getDescriptor)(t, selfKeys),
        psbtParams: (0, descriptor_2.getPsbtParams)(t),
        stages: defaultStagesCombinedAB,
    });
}
function describeCreatePsbtSelectTapLeafScript(t, selfDescriptor, selectTapLeafScript, tapLeafName) {
    const otherDescriptor = (0, descriptor_2.getDescriptor)('Wsh2Of3', otherKeys);
    const signKeys = selfKeys.filter((k) => (0, descriptor_2.containsKey)(selectTapLeafScript, k)).map((k) => toPlain(k));
    (0, assert_1.default)(signKeys.length === 1);
    describeCreatePsbt(`${t}-${tapLeafName}`, {
        descriptorSelf: selfDescriptor,
        psbt: (0, descriptor_2.mockPsbt)([
            { descriptor: selfDescriptor, index: 0, selectTapLeafScript },
            { descriptor: selfDescriptor, index: 1, id: { vout: 1 }, selectTapLeafScript },
        ], [
            {
                descriptor: otherDescriptor,
                index: 0,
                value: BigInt(4e5),
                external: true,
            },
            { descriptor: selfDescriptor, index: 0, value: BigInt(4e5) },
        ]),
        stages: [
            { name: 'unsigned', keys: [] },
            { name: 'signed', keys: signKeys },
        ],
    });
}
describeCreatePsbt2Of3('Wsh2Of3');
describeCreatePsbt2Of3('Wsh2Of3CltvDrop');
describeCreatePsbt2Of3('Tr2Of3-NoKeyPath');
describeCreatePsbt('Tr1Of3-NoKeyPath-Tree', {
    descriptorSelf: (0, descriptor_2.getDescriptor)('Tr1Of3-NoKeyPath-Tree', selfKeys),
    psbtParams: {},
    stages: getDefaultStagesSeparateAB(),
});
{
    const t = 'Tr1Of3-NoKeyPath-Tree-Plain';
    const selfDescriptor = (0, descriptor_2.getDescriptor)(t, selfKeys);
    (0, descriptor_2.getTapLeafScripts)(selfDescriptor).forEach((selectTapLeafScript, i) => {
        describeCreatePsbtSelectTapLeafScript(t, selfDescriptor, wasm_utxo_1.Miniscript.fromString(selectTapLeafScript, 'tap'), `TapLeafScript${i}`);
    });
}
describeCreatePsbt('Tr1Of3-NoKeyPath-Tree-PlainKeys', {
    descriptorSelf: (0, descriptor_2.getDescriptor)('Tr1Of3-NoKeyPath-Tree-Plain', selfKeys),
    psbtParams: {},
    stages: getDefaultStagesSeparateAB({ plain: true }),
});
{
    const descriptorSelf = (0, descriptor_2.getDescriptor)('Wsh2Of3', selfKeys);
    const descriptorOther = (0, descriptor_2.getDescriptor)('Wsh2Of3', otherKeys);
    describeCreatePsbt('Wsh2Of3-CustomInputSequence', {
        descriptorSelf,
        psbt: (0, descriptor_2.mockPsbt)([
            { descriptor: descriptorSelf, index: 0 },
            { descriptor: descriptorSelf, index: 1, id: { vout: 1 }, sequence: 123 },
        ], [
            {
                descriptor: descriptorOther,
                index: 0,
                value: BigInt(4e5),
                external: true,
            },
            { descriptor: descriptorSelf, index: 0, value: BigInt(4e5) },
        ]),
        stages: defaultStagesCombinedAB,
    });
}
