// @bitgo/utxo-core/dist/cjs/test/descriptor/psbt/VirtualSize.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const VirtualSize_1 = require("../../../src/descriptor/VirtualSize");
const descriptor_1 = require("../../../src/testutil/descriptor");
const testutil_1 = require("../../../src/testutil");
const descriptor_2 = require("../../../src/descriptor");
describe('VirtualSize', function () {
    describe('getInputVSizesForDescriptorWallet', function () {
        it('returns the input virtual sizes for a descriptor wallet', function () {
            assert_1.default.deepStrictEqual((0, VirtualSize_1.getInputVSizesForDescriptors)(new Map([
                ['foo', (0, descriptor_1.getDescriptor)('Wsh2Of2')],
                ['bar', (0, descriptor_1.getDescriptor)('Wsh2Of2')],
            ])), {
                foo: 96,
                bar: 96,
            });
            assert_1.default.deepStrictEqual((0, VirtualSize_1.getInputVSizesForDescriptors)((0, descriptor_1.getDescriptorMap)('Wsh2Of3')), {
                external: 105,
                internal: 105,
            });
        });
    });
    describe('getChangeOutputVSizesForDescriptor', function () {
        it('returns the output virtual sizes for a descriptor', function () {
            assert_1.default.deepStrictEqual((0, VirtualSize_1.getChangeOutputVSizesForDescriptor)((0, descriptor_1.getDescriptor)('Wsh2Of2')), {
                inputVSize: 96,
                outputVSize: 34,
            });
        });
    });
    function describeWithTemplate(t, inputSize, outputSize) {
        describe(`getVirtualSize ${t}`, function () {
            it('returns expected virtual size', function () {
                assert_1.default.deepStrictEqual((0, VirtualSize_1.getVirtualSize)({
                    inputs: [{ descriptorName: 'internal' }],
                    outputs: [{ script: Buffer.alloc(34) }],
                }, (0, descriptor_1.getDescriptorMap)(t)), inputSize + outputSize + 11);
                const descriptor = (0, descriptor_1.getDescriptor)(t);
                const nInputs = 10000;
                assert_1.default.deepStrictEqual((0, VirtualSize_1.getVirtualSize)({
                    /* as proof we can pass 10_000 inputs */
                    inputs: Array.from({ length: nInputs }).map(() => descriptor),
                    outputs: [],
                }), inputSize * nInputs + 11);
            });
        });
        describe(`getVirtualSizeForPsbt ${t}`, function () {
            const keys = (0, testutil_1.getKeyTriple)('a');
            const descriptorSelf = (0, descriptor_1.getDescriptor)(t, keys.map((k) => k.neutered().toBase58()));
            const descriptorOther = (0, descriptor_1.getDescriptor)(t, (0, descriptor_1.getDefaultXPubs)('b'));
            it('returns expected virtual size', function () {
                const psbt = (0, descriptor_1.mockPsbtDefault)({ descriptorSelf, descriptorOther });
                const descriptorMap = new Map([['internal', descriptorSelf]]);
                const expectedVirtualSize = inputSize * 2 + outputSize * 2 + 11;
                assert_1.default.deepStrictEqual((0, VirtualSize_1.getVirtualSizeEstimateForPsbt)(psbt, descriptorMap), expectedVirtualSize);
                psbt.signAllInputsHD(keys[0]);
                psbt.signAllInputsHD(keys[1]);
                (0, descriptor_2.finalizePsbt)(psbt);
                // TODO(BTC-1797): figure out why we overestimate by 1
                assert_1.default.strictEqual(psbt.extractTransaction().virtualSize(), expectedVirtualSize - 1);
            });
        });
    }
    describeWithTemplate('Wsh2Of3', 105, 43);
    describeWithTemplate('Tr2Of3-NoKeyPath', 109, 43);
});
