// @bitgo/sdk-coin-bsc/dist/src/bsc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bsc = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class Bsc extends abstract_eth_1.AbstractEthLikeNewCoins {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
    }
    static createInstance(bitgo, staticsCoin) {
        return new Bsc(bitgo, staticsCoin);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc */
    allowsAccountConsolidations() {
        return true;
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /**
     * Make a query to BSCScan for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @param {string} apiKey optional API key to use instead of the one from the environment
     * @returns {Promise<Object>} response from BSCScan
     */
    async recoveryBlockchainExplorerQuery(query, apiKey) {
        const apiToken = apiKey || sdk_core_1.common.Environments[this.bitgo.getEnv()].bscscanApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].bscscanBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    /**
     * Verify if a tss transaction is valid
     *
     * @param {VerifyEthTransactionOptions} params
     * @param {TransactionParams} params.txParams - params object passed to send
     * @param {TransactionPrebuild} params.txPrebuild - prebuild object returned by server
     * @param {Wallet} params.wallet - Wallet object to obtain keys to verify against
     * @returns {boolean}
     */
    async verifyTssTransaction(params) {
        const { txParams, txPrebuild, wallet } = params;
        if (!txParams?.recipients &&
            !(txParams.prebuildTx?.consolidateId ||
                (txParams.type && ['acceleration', 'fillNonce', 'transferToken', 'tokenApproval'].includes(txParams.type)))) {
            throw new Error(`missing txParams`);
        }
        if (!wallet || !txPrebuild) {
            throw new Error(`missing params`);
        }
        if (txParams.hop && txParams.recipients && txParams.recipients.length > 1) {
            throw new Error(`tx cannot be both a batch and hop transaction`);
        }
        return true;
    }
}
exports.Bsc = Bsc;
