// @bitgo/public-types/src/schema/lightning/payment.ts
import * as t from "io-ts";
import { BigIntFromString } from "io-ts-types";

export const LightningOnchainRecipient = t.type({
  amountSat: BigIntFromString,
  address: t.string,
});
export type LightningOnchainRecipient = t.TypeOf<
  typeof LightningOnchainRecipient
>;

export const LightningOnchainRequest = t.intersection([
  t.type({
    recipients: t.array(LightningOnchainRecipient),
  }),
  t.partial({
    satsPerVbyte: BigIntFromString,
    numBlocks: t.number,
  }),
]);
export type LightningOnchainRequest = t.TypeOf<typeof LightningOnchainRequest>;

export const LightningPaymentRequest = t.intersection([
  t.type({
    invoice: t.string,
  }),
  t.partial({
    amountMsat: BigIntFromString,
    feeLimitMsat: BigIntFromString,
    feeLimitRatio: t.number,
  }),
]);

export const LightningPaymentRequestWithSig = t.type({
  signedRequest: LightningPaymentRequest,
  signature: t.string,
});

export type LightningPaymentRequest = t.TypeOf<typeof LightningPaymentRequest>;
export type LightningPaymentRequestWithSig = t.TypeOf<
  typeof LightningPaymentRequestWithSig
>;
