// @bitgo/abstract-utxo/dist/esm/abstractUtxoCoin.js
import assert from 'assert';
import { randomBytes } from 'crypto';
import _ from 'lodash';
import * as utxolib from '@bitgo/utxo-lib';
import { BIP32, fixedScriptWallet } from '@bitgo/wasm-utxo';
import { bitgo, getMainnet } from '@bitgo/utxo-lib';
import { BaseCoin, InvalidAddressDerivationPropertyError, InvalidAddressError, isTriple, KeyIndices, multisigTypes, TxIntentMismatchError, TxIntentMismatchRecipientError, isValidPrv, isValidXprv, } from '@bitgo/sdk-core';
import { backupKeyRecovery, forCoin, recoverCrossChain, v1BackupKeyRecovery, v1Sweep, } from './recovery';
import { getReplayProtectionPubkeys, isReplayProtectionUnspent } from './transaction/fixedScript/replayProtection';
import { supportedCrossChainRecoveries } from './config';
import { assertValidTransactionRecipient, explainTx, fromExtendedAddressFormat, isScriptRecipient, parseTransaction, verifyTransaction, } from './transaction';
import { AggregateValidationError, ErrorMissingOutputs, ErrorImplicitExternalOutputs, } from './transaction/descriptor/verifyTransaction';
import { assertDescriptorWalletAddress, getDescriptorMapFromWallet, isDescriptorWallet } from './descriptor';
import { getFullNameFromCoinName, getMainnetCoinName, getNetworkFromCoinName, isTestnetCoin, isUtxoCoinNameMainnet, } from './names';
import { assertFixedScriptWalletAddress } from './address/fixedScript';
import { isSdkBackend } from './transaction/types';
import { decodePsbtWith, encodeTransaction, stringToBufferTryFormats } from './transaction/decode';
import { fetchKeychains, toBip32Triple } from './keychains';
import { verifyKeySignature, verifyUserPublicKey } from './verifyKey';
import { getPolicyForEnv } from './descriptor/validatePolicy';
import { signTransaction } from './transaction/signTransaction';
import { isUtxoWalletData } from './wallet';
import { isDescriptorWalletData } from './descriptor/descriptorWallet';
export class ErrorDeprecatedTxFormat extends Error {
    constructor(txFormat) {
        super(`SDK support for txFormat=${txFormat} is deprecated on this environment. Please use psbt instead.`);
    }
}
const { isChainCode, scriptTypeForChain, outputScripts } = bitgo;
/**
 * Check if a decoded transaction has at least one taproot key path spend (MuSig2) input.
 * Works for both utxolib UtxoPsbt and wasm-utxo BitGoPsbt.
 */
function hasKeyPathSpendInput(tx, pubs, coinName) {
    if (tx instanceof bitgo.UtxoPsbt) {
        return bitgo.isTransactionWithKeyPathSpendInput(tx);
    }
    if (tx instanceof fixedScriptWallet.BitGoPsbt) {
        assert(pubs && isTriple(pubs), 'pub triple is required to check for key path spend inputs in wasm-utxo PSBT');
        const rootWalletKeys = fixedScriptWallet.RootWalletKeys.fromXpubs(pubs);
        const replayProtection = { publicKeys: getReplayProtectionPubkeys(coinName) };
        const parsed = tx.parseTransactionWithWalletKeys(rootWalletKeys, { replayProtection });
        return parsed.inputs.some((input) => input.scriptType === 'p2trMusig2KeyPath');
    }
    return false;
}
/**
 * Convert ValidationError to TxIntentMismatchRecipientError with structured data
 *
 * This preserves the structured error information from the original ValidationError
 * by extracting the mismatched outputs and converting them to the standardized format.
 * The original error is preserved as the `cause` for debugging purposes.
 */
function convertValidationErrorToTxIntentMismatch(error, reqId, txParams, txHex, txExplanation) {
    const mismatchedRecipients = [];
    for (const err of error.errors) {
        if (err instanceof ErrorMissingOutputs) {
            mismatchedRecipients.push(...err.missingOutputs.map((output) => ({
                address: output.address,
                amount: output.amount.toString(),
            })));
        }
        else if (err instanceof ErrorImplicitExternalOutputs) {
            mismatchedRecipients.push(...err.implicitExternalOutputs.map((output) => ({
                address: output.address,
                amount: output.amount.toString(),
            })));
        }
    }
    const txIntentError = new TxIntentMismatchRecipientError(error.message, reqId, [txParams], txHex, mismatchedRecipients, txExplanation);
    // Preserve the original structured error as the cause for debugging
    // See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause
    txIntentError.cause = error;
    return txIntentError;
}
export class AbstractUtxoCoin extends BaseCoin {
    constructor(bitgo, amountType = 'number') {
        super(bitgo);
        this.amountType = amountType;
    }
    get defaultSdkBackend() {
        return isUtxoCoinNameMainnet(this.name) ? 'utxolib' : 'wasm-utxo';
    }
    /**
     * @deprecated - will be removed when we drop support for utxolib
     * Use `name` property instead.
     */
    get network() {
        return getNetworkFromCoinName(this.name);
    }
    getChain() {
        return this.name;
    }
    getFamily() {
        return getMainnetCoinName(this.name);
    }
    getFullName() {
        return getFullNameFromCoinName(this.name);
    }
    /** Indicates whether the coin supports a block target */
    supportsBlockTarget() {
        // FIXME: the SDK does not seem to use this anywhere so it is unclear what the purpose of this method is
        switch (getMainnet(this.network)) {
            case utxolib.networks.bitcoin:
            case utxolib.networks.dogecoin:
                return true;
            default:
                return false;
        }
    }
    sweepWithSendMany() {
        return true;
    }
    /** @deprecated */
    static get validAddressTypes() {
        return [...outputScripts.scriptTypes2Of3];
    }
    /**
     * Returns the factor between the base unit and its smallest subdivison
     * @return {number}
     */
    getBaseFactor() {
        return 1e8;
    }
    /**
     * Check if an address is valid
     * @param address
     * @param param
     */
    isValidAddress(address, param) {
        if (typeof param === 'boolean' && param) {
            throw new Error('deprecated');
        }
        // By default, allow all address formats.
        // At the time of writing, the only additional address format is bch cashaddr.
        const anyFormat = param?.anyFormat ?? true;
        try {
            // Find out if the address is valid for any format. Tries all supported formats by default.
            // Throws if address cannot be decoded with any format.
            const [format, script] = utxolib.addressFormat.toOutputScriptAndFormat(address, this.network);
            // unless anyFormat is set, only 'default' is allowed.
            if (!anyFormat && format !== 'default') {
                return false;
            }
            // make sure that address is in normal representation for given format.
            return address === utxolib.addressFormat.fromOutputScriptWithFormat(script, format, this.network);
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Return boolean indicating whether input is valid public key for the coin.
     *
     * @param {String} pub the pub to be checked
     * @returns {Boolean} is it valid?
     */
    isValidPub(pub) {
        try {
            return BIP32.fromBase58(pub).isNeutered();
        }
        catch (e) {
            return false;
        }
    }
    /**
     * This is called before crafting the HTTP request to the BitGo API.
     * It converts the recipient address `scriptPubKey:...` to { script: string } | { address: string }.
     */
    preprocessBuildParams(params) {
        if (params.recipients !== undefined) {
            params.recipients =
                params.recipients instanceof Array
                    ? params?.recipients?.map((recipient) => {
                        const { address, ...rest } = recipient;
                        return { ...rest, ...fromExtendedAddressFormat(address) };
                    })
                    : params.recipients;
        }
        return params;
    }
    /**
     * Get the latest block height
     * @param reqId
     */
    async getLatestBlockHeight(reqId) {
        if (reqId) {
            this.bitgo.setRequestTracer(reqId);
        }
        const chainhead = await this.bitgo.get(this.url('/public/block/latest')).result();
        return chainhead.height;
    }
    checkRecipient(recipient) {
        assertValidTransactionRecipient(recipient);
        if (!isScriptRecipient(recipient.address)) {
            super.checkRecipient(recipient);
        }
    }
    /**
     * Run custom coin logic after a transaction prebuild has been received from BitGo
     * @param prebuild
     */
    async postProcessPrebuild(prebuild) {
        const tx = this.decodeTransactionFromPrebuild(prebuild);
        if (_.isUndefined(prebuild.blockHeight)) {
            prebuild.blockHeight = (await this.getLatestBlockHeight());
        }
        return _.extend({}, prebuild, { txHex: encodeTransaction(tx).toString('hex') });
    }
    /**
     * @deprecated - will be removed when we drop support for utxolib
     * Determine an address' type based on its witness and redeem script presence
     * @param addressDetails
     */
    static inferAddressType(addressDetails) {
        return isChainCode(addressDetails.chain) ? scriptTypeForChain(addressDetails.chain) : null;
    }
    createTransactionFromHex(hex) {
        return utxolib.bitgo.createTransactionFromHex(hex, this.network, this.amountType);
    }
    decodeTransaction(input, decodeWith = this.defaultSdkBackend) {
        if (typeof input === 'string') {
            const buffer = stringToBufferTryFormats(input, ['hex', 'base64']);
            return this.decodeTransaction(buffer, decodeWith);
        }
        if (utxolib.bitgo.isPsbt(input)) {
            return decodePsbtWith(input, this.name, decodeWith);
        }
        else {
            if (decodeWith !== 'utxolib') {
                console.error('received decodeWith hint %s, ignoring for legacy transaction', decodeWith);
            }
            return utxolib.bitgo.createTransactionFromBuffer(input, this.network, {
                amountType: this.amountType,
            });
        }
    }
    decodeTransactionAsPsbt(input) {
        const decoded = this.decodeTransaction(input);
        if (decoded instanceof fixedScriptWallet.BitGoPsbt || decoded instanceof utxolib.bitgo.UtxoPsbt) {
            return decoded;
        }
        throw new Error('expected psbt but got transaction');
    }
    decodeTransactionFromPrebuild(prebuild) {
        const string = prebuild.txHex ?? prebuild.txBase64;
        if (!string) {
            throw new Error('missing required txHex or txBase64 property');
        }
        let { decodeWith } = prebuild;
        if (decodeWith !== undefined) {
            if (typeof decodeWith !== 'string' || !isSdkBackend(decodeWith)) {
                console.error('decodeWith %s is not a valid value, using default', decodeWith);
                decodeWith = undefined;
            }
        }
        return this.decodeTransaction(string, decodeWith);
    }
    /**
     * Extract and fill transaction details such as internal/change spend, external spend (explicit vs. implicit), etc.
     * @param params
     * @returns {*}
     */
    async parseTransaction(params) {
        return parseTransaction(this, params);
    }
    /**
     * @deprecated - use function verifyUserPublicKey instead
     */
    verifyUserPublicKey(params) {
        return verifyUserPublicKey(this.bitgo, params);
    }
    /**
     * @deprecated - use function verifyKeySignature instead
     */
    verifyKeySignature(params) {
        return verifyKeySignature(params);
    }
    /**
     * Verify that a transaction prebuild complies with the original intention
     *
     * @param params
     * @param params.txParams params object passed to send
     * @param params.txPrebuild prebuild object returned by server
     * @param params.txPrebuild.txHex prebuilt transaction's txHex form
     * @param params.wallet Wallet object to obtain keys to verify against
     * @param params.verification Object specifying some verification parameters
     * @param params.verification.disableNetworking Disallow fetching any data from the internet for verification purposes
     * @param params.verification.keychains Pass keychains manually rather than fetching them by id
     * @param params.verification.addresses Address details to pass in for out-of-band verification
     * @returns {boolean} True if verification passes
     * @throws {TxIntentMismatchError} if transaction validation fails
     * @throws {TxIntentMismatchRecipientError} if transaction recipients don't match user intent
     */
    async verifyTransaction(params) {
        try {
            return await verifyTransaction(this, this.bitgo, params);
        }
        catch (error) {
            if (error instanceof AggregateValidationError) {
                const txExplanation = await TxIntentMismatchError.tryGetTxExplanation(this, params.txPrebuild);
                throw convertValidationErrorToTxIntentMismatch(error, params.reqId, params.txParams, params.txPrebuild.txHex, txExplanation);
            }
            throw error;
        }
    }
    /**
     * Make sure an address is valid and throw an error if it's not.
     * @param params.address The address string on the network
     * @param params.addressType
     * @param params.keychains Keychain objects with xpubs
     * @param params.coinSpecific Coin-specific details for the address such as a witness script
     * @param params.chain Derivation chain
     * @param params.index Derivation index
     * @throws {InvalidAddressError}
     * @throws {InvalidAddressDerivationPropertyError}
     * @throws {UnexpectedAddressError}
     */
    async isWalletAddress(params, wallet) {
        const { address, keychains, chain, index } = params;
        if (!this.isValidAddress(address)) {
            throw new InvalidAddressError(`invalid address: ${address}`);
        }
        if (wallet && isDescriptorWallet(wallet)) {
            if (!keychains) {
                throw new Error('missing required param keychains');
            }
            if (!isTriple(keychains)) {
                throw new Error('keychains must be a triple');
            }
            assertDescriptorWalletAddress(this.name, params, getDescriptorMapFromWallet(wallet, toBip32Triple(keychains), getPolicyForEnv(this.bitgo.env)));
            return true;
        }
        if ((_.isUndefined(chain) && _.isUndefined(index)) || !(_.isFinite(chain) && _.isFinite(index))) {
            throw new InvalidAddressDerivationPropertyError(`address validation failure: invalid chain (${chain}) or index (${index})`);
        }
        if (!keychains) {
            throw new Error('missing required param keychains');
        }
        assertFixedScriptWalletAddress(this.name, {
            address,
            keychains,
            format: params.format ?? 'base58',
            chain,
            index,
        });
        return true;
    }
    /**
     * @param addressType
     * @returns true iff coin supports spending from unspentType
     */
    supportsAddressType(addressType) {
        return utxolib.bitgo.outputScripts.isSupportedScriptType(this.network, addressType);
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return multisigTypes.onchain;
    }
    /**
     * @param chain
     * @return true iff coin supports spending from chain
     */
    supportsAddressChain(chain) {
        return isChainCode(chain) && this.supportsAddressType(utxolib.bitgo.scriptTypeForChain(chain));
    }
    keyIdsForSigning() {
        return [KeyIndices.USER, KeyIndices.BACKUP, KeyIndices.BITGO];
    }
    async getMusig2Nonces(psbt, walletId) {
        const buffer = encodeTransaction(psbt);
        const response = await this.bitgo
            .post(this.url('/wallet/' + walletId + '/tx/signpsbt'))
            .send({ psbt: buffer.toString('hex') })
            .result();
        if (psbt instanceof utxolib.bitgo.UtxoPsbt) {
            return decodePsbtWith(response.psbt, this.name, 'utxolib');
        }
        else {
            return decodePsbtWith(response.psbt, this.name, 'wasm-utxo');
        }
    }
    /**
     * @deprecated Use getMusig2Nonces instead
     * @returns input psbt added with deterministic MuSig2 nonce for bitgo key for each MuSig2 inputs.
     * @param psbtHex all MuSig2 inputs should contain user MuSig2 nonce
     * @param walletId
     */
    async signPsbt(psbtHex, walletId) {
        const psbt = await this.getMusig2Nonces(this.decodeTransactionAsPsbt(psbtHex), walletId);
        return { psbt: encodeTransaction(psbt).toString('hex') };
    }
    /**
     * @returns input psbt added with deterministic MuSig2 nonce for bitgo key for each MuSig2 inputs from OVC.
     * @param ovcJson JSON object provided by OVC with fields psbtHex and walletId
     */
    async signPsbtFromOVC(ovcJson) {
        assert(ovcJson['psbtHex'], 'ovcJson must contain psbtHex');
        assert(ovcJson['walletId'], 'ovcJson must contain walletId');
        const hex = ovcJson['psbtHex'];
        const walletId = ovcJson['walletId'];
        const psbt = await this.getMusig2Nonces(this.decodeTransactionAsPsbt(hex), walletId);
        return _.extend(ovcJson, { txHex: encodeTransaction(psbt).toString('hex') });
    }
    /**
     * Assemble keychain and half-sign prebuilt transaction
     * @param params - {@see SignTransactionOptions}
     * @returns {Promise<SignedTransaction | HalfSignedUtxoTransaction>}
     */
    async signTransaction(params) {
        return signTransaction(this, this.bitgo, params);
    }
    /**
     * Sign a transaction with a custom signing function. Example use case is express external signer
     * @param customSigningFunction custom signing function that returns a single signed transaction
     * @param signTransactionParams parameters for custom signing function. Includes txPrebuild and pubs
     *
     * @returns signed transaction as hex string
     */
    async signWithCustomSigningFunction(customSigningFunction, signTransactionParams) {
        const txHex = signTransactionParams.txPrebuild.txHex;
        assert(txHex, 'missing txHex parameter');
        const tx = this.decodeTransaction(txHex);
        const isTxWithKeyPathSpendInput = hasKeyPathSpendInput(tx, signTransactionParams.pubs, this.name);
        if (!isTxWithKeyPathSpendInput) {
            return await customSigningFunction({ ...signTransactionParams, coin: this });
        }
        const getTxHex = (v) => {
            if ('txHex' in v) {
                return v.txHex;
            }
            throw new Error('txHex not found in signTransaction result');
        };
        const signerNonceTx = await customSigningFunction({
            ...signTransactionParams,
            signingStep: 'signerNonce',
            coin: this,
        });
        const { pubs } = signTransactionParams;
        assert(pubs === undefined || isTriple(pubs));
        const cosignerNonceTx = await this.signTransaction({
            ...signTransactionParams,
            pubs,
            txPrebuild: { ...signTransactionParams.txPrebuild, txHex: getTxHex(signerNonceTx) },
            signingStep: 'cosignerNonce',
        });
        return await customSigningFunction({
            ...signTransactionParams,
            txPrebuild: { ...signTransactionParams.txPrebuild, txHex: getTxHex(cosignerNonceTx) },
            signingStep: 'signerSignature',
            coin: this,
        });
    }
    /**
     * @param unspent
     * @returns {boolean}
     */
    isBitGoTaintedUnspent(unspent) {
        return isReplayProtectionUnspent(unspent, this.name);
    }
    /**
     * Decompose a raw psbt/transaction into useful information, such as the total amounts,
     * change amounts, and transaction outputs.
     * @param params
     */
    async explainTransaction(params, wallet) {
        if (!params.pubs) {
            // if no pubs are provided, opportunistically fetch the keychains from the wallet
            if (wallet) {
                const keychains = await fetchKeychains(this, wallet);
                params.pubs = toBip32Triple(keychains).map((k) => k.neutered().toBase58());
            }
        }
        return explainTx(this.decodeTransactionFromPrebuild(params), params, this.name);
    }
    /**
     * @deprecated - use {@see backupKeyRecovery}
     * Builds a funds recovery transaction without BitGo
     * @param params - {@see backupKeyRecovery}
     */
    async recover(params) {
        return backupKeyRecovery(this, this.bitgo, params);
    }
    async recoverV1(params) {
        return v1BackupKeyRecovery(this, this.bitgo, params);
    }
    async sweepV1(params) {
        return v1Sweep(this, this.bitgo, params);
    }
    /**
     * Recover coin that was sent to wrong chain
     * @param params
     * @param params.txid The txid of the faulty transaction
     * @param params.recoveryAddress address to send recovered funds to
     * @param params.wallet the wallet that received the funds
     * @param params.recoveryCoin the coin type of the wallet that received the funds
     * @param params.signed return a half-signed transaction (default=true)
     * @param params.walletPassphrase the wallet passphrase
     * @param params.xprv the unencrypted xprv (used instead of wallet passphrase)
     * @param params.apiKey for utxo coins other than [BTC,TBTC] this is a Block Chair api key
     * @returns {*}
     */
    async recoverFromWrongChain(params) {
        const { txid, recoveryAddress, wallet, walletPassphrase, xprv, apiKey } = params;
        // params.recoveryCoin used to be params.coin, backwards compatibility
        const recoveryCoin = params.coin || params.recoveryCoin;
        if (!recoveryCoin) {
            throw new Error('missing required object recoveryCoin');
        }
        // signed should default to true, and only be disabled if explicitly set to false (not undefined)
        const signed = params.signed !== false;
        const sourceCoinFamily = this.getFamily();
        const recoveryCoinFamily = recoveryCoin.getFamily();
        const supportedRecoveryCoins = supportedCrossChainRecoveries[sourceCoinFamily];
        if (_.isUndefined(supportedRecoveryCoins) || !supportedRecoveryCoins.includes(recoveryCoinFamily)) {
            throw new Error(`Recovery of ${sourceCoinFamily} balances from ${recoveryCoinFamily} wallets is not supported.`);
        }
        return await recoverCrossChain(this.bitgo, {
            sourceCoin: this,
            recoveryCoin,
            walletId: wallet,
            txid,
            recoveryAddress,
            walletPassphrase: signed ? walletPassphrase : undefined,
            xprv: signed ? xprv : undefined,
            apiKey,
        });
    }
    /**
     * Generate bip32 key pair
     *
     * @param seed
     * @returns {Object} object with generated pub and prv
     */
    generateKeyPair(seed) {
        if (!seed) {
            // An extended private key has both a normal 256 bit private key and a 256
            // bit chain code, both of which must be random. 512 bits is therefore the
            // maximum entropy and gives us maximum security against cracking.
            seed = randomBytes(512 / 8);
        }
        const extendedKey = BIP32.fromSeed(seed);
        return {
            pub: extendedKey.neutered().toBase58(),
            prv: extendedKey.toBase58(),
        };
    }
    /**
     * Determines the default transaction format based on wallet properties and network
     * @param wallet - The wallet to check
     * @param requestedFormat - Optional explicitly requested format
     * @returns The transaction format to use, or undefined if no default applies
     */
    getDefaultTxFormat(wallet, requestedFormat) {
        // If format is explicitly requested, use it
        if (requestedFormat !== undefined) {
            if (isTestnetCoin(this.name) && requestedFormat === 'legacy') {
                throw new ErrorDeprecatedTxFormat(requestedFormat);
            }
            return requestedFormat;
        }
        return 'psbt-lite';
    }
    async getExtraPrebuildParams(buildParams) {
        const txFormat = this.getDefaultTxFormat(buildParams.wallet, buildParams.txFormat);
        let changeAddressType = buildParams.changeAddressType;
        // if the addressType is not specified, we need to default to p2trMusig2 for testnet hot wallets for staged rollout of p2trMusig2
        if (buildParams.addressType === undefined && // addressType is deprecated and replaced by `changeAddress`
            buildParams.changeAddressType === undefined &&
            buildParams.changeAddress === undefined &&
            buildParams.wallet.type() === 'hot') {
            changeAddressType = ['p2trMusig2', 'p2wsh', 'p2shP2wsh', 'p2sh', 'p2tr'];
        }
        return {
            txFormat,
            changeAddressType,
        };
    }
    preCreateBitGo(params) {
        return;
    }
    async presignTransaction(params) {
        if (params.walletData && isUtxoWalletData(params.walletData) && isDescriptorWalletData(params.walletData)) {
            return params;
        }
        // In the case that we have a 'psbt-lite' transaction format, we want to indicate in signing to not fail
        const txHex = (params.txHex ?? params.txPrebuild?.txHex);
        if (txHex &&
            utxolib.bitgo.isPsbt(txHex) &&
            utxolib.bitgo.isPsbtLite(utxolib.bitgo.createPsbtFromHex(txHex, this.network)) &&
            params.allowNonSegwitSigningWithoutPrevTx === undefined) {
            return { ...params, allowNonSegwitSigningWithoutPrevTx: true };
        }
        return params;
    }
    async supplementGenerateWallet(walletParams, keychains) {
        return walletParams;
    }
    transactionDataAllowed() {
        return false;
    }
    valuelessTransferAllowed() {
        return false;
    }
    getRecoveryProvider(apiToken) {
        return forCoin(this.getChain(), apiToken);
    }
    /** @inheritDoc */
    auditDecryptedKey({ multiSigType, publicKey, prv, }) {
        if (multiSigType === 'tss') {
            throw new Error('tss auditing is not supported for this coin');
        }
        if (!isValidPrv(prv) && !isValidXprv(prv)) {
            throw new Error('invalid private key');
        }
        if (publicKey) {
            const genPubKey = BIP32.fromBase58(prv).neutered().toBase58();
            if (genPubKey !== publicKey) {
                throw new Error('public key does not match private key');
            }
        }
    }
}
