// @bitgo/abstract-utxo/dist/esm/wasmUtil.js
import { BIP32, ECPair, Psbt } from '@bitgo/wasm-utxo';
import * as utxolib from '@bitgo/utxo-lib';
/**
 * Convert a utxo-lib BIP32Interface to a wasm-utxo BIP32 instance.
 * Preserves private key by using base58 serialization.
 */
export function toWasmBIP32(key) {
    if (key instanceof BIP32) {
        return key;
    }
    // All utxo-lib BIP32Interface instances have toBase58
    return BIP32.fromBase58(key.toBase58());
}
/**
 * Convert a wasm-utxo BIP32 to a utxo-lib BIP32Interface.
 * Used at boundaries where utxo-lib APIs require their own BIP32Interface type.
 */
export function toUtxolibBIP32(key) {
    return utxolib.bip32.fromBase58(key.toBase58());
}
export function toWasmECPair(key) {
    if (key instanceof ECPair) {
        return key;
    }
    if (key instanceof Uint8Array) {
        return ECPair.from(key);
    }
    if (key.privateKey) {
        return ECPair.fromPrivateKey(key.privateKey);
    }
    return ECPair.fromPublicKey(key.publicKey);
}
export function isUtxoLibPsbt(psbt) {
    return psbt instanceof utxolib.Psbt || psbt instanceof utxolib.bitgo.UtxoPsbt;
}
export function toWasmPsbt(psbt) {
    if (psbt instanceof Psbt) {
        return psbt;
    }
    if (psbt instanceof Uint8Array) {
        return Psbt.deserialize(psbt);
    }
    if (isUtxoLibPsbt(psbt)) {
        return Psbt.deserialize(psbt.toBuffer());
    }
    throw new Error('Unsupported PSBT type');
}
/**
 * Sum the `value` property of an array of objects.
 */
export function sumValues(arr) {
    return arr.reduce((sum, e) => sum + e.value, BigInt(0));
}
