// @bitgo/sdk-hmac/dist/src/hmac.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateHMAC = calculateHMAC;
exports.calculateHMACSubject = calculateHMACSubject;
exports.calculateRequestHMAC = calculateRequestHMAC;
exports.calculateRequestHeaders = calculateRequestHeaders;
exports.verifyResponse = verifyResponse;
const urlLib = __importStar(require("url"));
const sjcl = __importStar(require("@bitgo/sjcl"));
const util_1 = require("./util");
/**
 * Calculate the HMAC for the given key and message
 * @param key {String} - the key to use for the HMAC
 * @param message {String} - the actual message to HMAC
 * @returns {*} - the result of the HMAC operation
 */
function calculateHMAC(key, message) {
    return (0, util_1.createHmacWithSha256)(key, message);
}
/**
 * Calculate the subject string that is to be HMAC'ed for a HTTP request or response
 * @param urlPath request url, including query params
 * @param text request body text
 * @param timestamp request timestamp from `Date.now()`
 * @param statusCode Only set for HTTP responses, leave blank for requests
 * @param method request method
 * @param authVersion authentication version (2 or 3)
 * @param useOriginalPath whether to use the original urlPath without parsing (default false)
 * @returns {string | Buffer}
 */
function calculateHMACSubject({ urlPath, text, timestamp, statusCode, method, authVersion }, useOriginalPath = false) {
    /* Normalize legacy 'del' to 'delete' for backward compatibility */
    if (method === 'del') {
        method = 'delete';
    }
    let queryPath = urlPath;
    if (!useOriginalPath) {
        const urlDetails = urlLib.parse(urlPath);
        queryPath = urlDetails.query && urlDetails.query.length > 0 ? urlDetails.path : urlDetails.pathname;
    }
    let prefixedText;
    if (statusCode !== undefined && isFinite(statusCode) && Number.isInteger(statusCode)) {
        prefixedText =
            authVersion === 3
                ? [method.toUpperCase(), timestamp, queryPath, statusCode].join('|')
                : [timestamp, queryPath, statusCode].join('|');
    }
    else {
        prefixedText =
            authVersion === 3
                ? [method.toUpperCase(), timestamp, '3.0', queryPath].join('|')
                : [timestamp, queryPath].join('|');
    }
    const isBuffer = Buffer.isBuffer(text);
    if (isBuffer) {
        return Buffer.concat([Buffer.from(prefixedText + '|', 'utf-8'), text]);
    }
    return [prefixedText, text].join('|');
}
/**
 * Calculate the HMAC for an HTTP request
 */
function calculateRequestHMAC({ url: urlPath, text, timestamp, token, method, authVersion }, useOriginalPath = false) {
    const signatureSubject = calculateHMACSubject({ urlPath, text, timestamp, method, authVersion }, useOriginalPath);
    // calculate the HMAC
    return calculateHMAC(token, signatureSubject);
}
/**
 * Calculate request headers with HMAC
 */
function calculateRequestHeaders({ url, text, token, method, authVersion }, useOriginalPath = false) {
    const timestamp = Date.now();
    const hmac = calculateRequestHMAC({ url, text, timestamp, token, method, authVersion }, useOriginalPath);
    // calculate the SHA256 hash of the token
    const hashDigest = sjcl.hash.sha256.hash(token);
    const tokenHash = sjcl.codec.hex.fromBits(hashDigest);
    return {
        hmac,
        timestamp,
        tokenHash,
    };
}
/**
 * Verify the HMAC for an HTTP response
 */
function verifyResponse({ url: urlPath, statusCode, text, timestamp, token, hmac, method, authVersion }, useOriginalPath = false) {
    const signatureSubject = calculateHMACSubject({
        urlPath,
        text,
        timestamp,
        statusCode,
        method,
        authVersion,
    }, useOriginalPath);
    // calculate the HMAC
    const expectedHmac = calculateHMAC(token, signatureSubject);
    // determine if the response is still within the validity window (5-minute backwards window, 1-minute forward window)
    const now = Date.now();
    const backwardValidityWindow = 1000 * 60 * 5;
    const forwardValidityWindow = 1000 * 60;
    const isInResponseValidityWindow = timestamp >= now - backwardValidityWindow && timestamp <= now + forwardValidityWindow;
    // verify the HMAC and timestamp
    return {
        isValid: expectedHmac === hmac,
        expectedHmac,
        signatureSubject,
        isInResponseValidityWindow,
        verificationTime: now,
    };
}
