// @bitgo/sdk-core/dist/src/bitgo/trading/tradingAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradingAccount = void 0;
const network_1 = require("./network");
class TradingAccount {
    constructor(enterpriseId, wallet, bitgo) {
        this.enterpriseId = enterpriseId;
        this.wallet = wallet;
        this.bitgo = bitgo;
    }
    get id() {
        return this.wallet.id();
    }
    /**
     * Signs an arbitrary payload with the user key on this trading account
     * @param params
     * @param params.payload arbitrary payload object (string | Record<string, unknown>)
     * @param params.walletPassphrase passphrase on this trading account, used to unlock the account user key
     * @returns hex-encoded signature of the payload
     */
    async signPayload(params) {
        const key = (await this.wallet.baseCoin.keychains().get({ id: this.wallet.keyIds()[0] }));
        const prv = this.wallet.bitgo.decrypt({
            input: key.encryptedPrv,
            password: params.walletPassphrase,
        });
        const payload = typeof params.payload === 'string' ? params.payload : JSON.stringify(params.payload);
        return (await this.wallet.baseCoin.signMessage({ prv }, payload)).toString('hex');
    }
    /**
     * Get Trade Network
     * To enable Off Exchange Allocation & Settlement, contact support@bitgo.com.
     * BitGo provides a UI experience for clients at: https://app.bitgo.com/web/enterprises/<enterpriseId>/allocate
     */
    toNetwork() {
        return new network_1.TradingNetwork(this.enterpriseId, this.wallet, this.bitgo);
    }
}
exports.TradingAccount = TradingAccount;
