// @bitgo/sdk-coin-polyx/dist/test/unit/polyx.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const src_1 = require("../../src");
const constants_1 = require("../../src/lib/constants");
const sinon = __importStar(require("sinon"));
const testData = __importStar(require("../resources/wrwUsers"));
const mocha_1 = require("mocha");
const resources_1 = require("../resources");
describe('Polyx:', function () {
    let bitgo;
    let baseCoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('polyx', src_1.Polyx.createInstance);
        bitgo.safeRegister('tpolyx', src_1.Tpolyx.createInstance);
        bitgo.initializeTestVars();
        baseCoin = bitgo.coin('tpolyx');
    });
    describe('Address Format Constants', function () {
        it('should have the correct address format constants', function () {
            // Verify the constants are defined correctly
            constants_1.POLYX_ADDRESS_FORMAT.should.equal(12);
            constants_1.TPOLYX_ADDRESS_FORMAT.should.equal(42);
        });
    });
    describe('Recover Transactions:', function () {
        const sandBox = sinon.createSandbox();
        const recoveryDestination = '5H56f31hSYGCRV3URjQHv2Cc4ZSkJNHTM8MKGtkkV6hzCqN7';
        const nonce = 0;
        let accountInfoCB;
        let headerInfoCB;
        let getFeeCB;
        beforeEach(function () {
            accountInfoCB = sandBox.stub(src_1.Polyx.prototype, 'getAccountInfo');
            headerInfoCB = sandBox.stub(src_1.Polyx.prototype, 'getHeaderInfo');
            getFeeCB = sandBox.stub(src_1.Polyx.prototype, 'getFee');
        });
        (0, mocha_1.afterEach)(function () {
            sandBox.restore();
        });
        it('should generate unsigned sweep correctly', async function () {
            accountInfoCB
                .withArgs(testData.unsignedSweepUser.walletAddress)
                .resolves({ nonce: 0, freeBalance: 100007000000 });
            headerInfoCB.resolves({
                headerNumber: testData.testnetBlock.blockNumber,
                headerHash: testData.testnetBlock.hash,
            });
            getFeeCB.withArgs(recoveryDestination, testData.unsignedSweepUser.walletAddress, 100007000000).resolves(74401);
            const commonKeyChain = testData.unsignedSweepUser.bitgoKey;
            const unsigned = await baseCoin.recover({ bitgoKey: commonKeyChain, recoveryDestination });
            unsigned.txRequests.should.not.be.undefined();
            unsigned.txRequests.length.should.equal(1);
            unsigned.txRequests[0].transactions.length.should.equal(1);
            unsigned.txRequests[0].walletCoin.should.equal('tpolyx');
            unsigned.txRequests[0].transactions[0].unsignedTx.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.serializedTx.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.scanIndex.should.equal(0);
            unsigned.txRequests[0].transactions[0].unsignedTx.coin.should.equal('tpolyx');
            unsigned.txRequests[0].transactions[0].unsignedTx.signableHex.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.derivationPath.should.equal('m/0');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs.length.should.equal(1);
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs[0].address.should.equal('5DnZQqbxtB3CWkLgfA6wpqdkCd9Bq7AX49RpLPJwW9mzoT7s');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs[0].valueString.should.equal('100006255990');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.inputs[0].value.should.equal(100006255990);
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs.length.should.equal(1);
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs[0].address.should.equal('5H56f31hSYGCRV3URjQHv2Cc4ZSkJNHTM8MKGtkkV6hzCqN7');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs[0].valueString.should.equal('100006255990');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.outputs[0].coinName.should.equal('tpolyx');
            unsigned.txRequests[0].transactions[0].unsignedTx.parsedTx.type.should.equal('');
            unsigned.txRequests[0].transactions[0].unsignedTx.feeInfo.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.feeInfo.fee.should.equal(0);
            unsigned.txRequests[0].transactions[0].unsignedTx.feeInfo.feeString.should.equal('0');
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.firstValid.should.not.be.undefined();
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.maxDuration.should.equal(2400);
            unsigned.txRequests[0].transactions[0].unsignedTx.coinSpecific.commonKeychain.should.equal('97d9da8bf544d07f3c7ae8ee06bb12a63f74478d0c461201ab94421a9c81d5b379553f0332daaa85cb8c65e1fe7e6907c62268c00812b08a4e1361178e304ddb');
        });
        it('should recover a tx for non-bitgo recoveries', async function () {
            accountInfoCB.withArgs(testData.wrwUser.walletAddress).resolves({ nonce: 0, freeBalance: 100007000000 });
            headerInfoCB.resolves({
                headerNumber: testData.testnetBlock.blockNumber,
                headerHash: testData.testnetBlock.hash,
            });
            getFeeCB.withArgs(recoveryDestination, testData.wrwUser.walletAddress, 100007000000).resolves(74401);
            const res = await baseCoin.recover({
                userKey: testData.wrwUser.userKey,
                backupKey: testData.wrwUser.backupKey,
                bitgoKey: testData.wrwUser.bitgoKey,
                walletPassphrase: testData.wrwUser.walletPassphrase,
                recoveryDestination: recoveryDestination,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            res.should.hasOwnProperty('scanIndex');
            const txBuilder = baseCoin.getBuilder().from(res.serializedTx);
            txBuilder
                .validity({
                firstValid: testData.testnetBlock.blockNumber,
                maxDuration: baseCoin.SWEEP_TXN_DURATION,
            })
                .referenceBlock(testData.testnetBlock.hash);
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.deepEqual(txJson.sender, testData.wrwUser.walletAddress);
            should_1.default.deepEqual(txJson.blockNumber, testData.testnetBlock.blockNumber);
            should_1.default.deepEqual(txJson.referenceBlock, testData.testnetBlock.hash);
            should_1.default.deepEqual(txJson.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson.nonce, nonce);
            should_1.default.deepEqual(txJson.tip, 0);
            should_1.default.deepEqual(txJson.transactionVersion, resources_1.txVersion);
            should_1.default.deepEqual(txJson.eraPeriod, baseCoin.SWEEP_TXN_DURATION);
        });
        it('should recover a txn for unsigned-sweep recoveries', async function () {
            accountInfoCB
                .withArgs(testData.unsignedSweepUser.walletAddress)
                .resolves({ nonce: 0, freeBalance: 100007000000 });
            headerInfoCB.resolves({
                headerNumber: testData.testnetBlock.blockNumber,
                headerHash: testData.testnetBlock.hash,
            });
            getFeeCB.withArgs(recoveryDestination, testData.unsignedSweepUser.walletAddress, 100007000000).resolves(74401);
            const res = await baseCoin.recover({
                bitgoKey: testData.unsignedSweepUser.bitgoKey,
                recoveryDestination: recoveryDestination,
            });
            res.should.not.be.empty();
            res.txRequests[0].transactions[0].unsignedTx.should.hasOwnProperty('serializedTx');
            res.txRequests[0].transactions[0].unsignedTx.should.hasOwnProperty('scanIndex');
            const txBuilder = baseCoin.getBuilder().from(res.txRequests[0].transactions[0].unsignedTx.serializedTx);
            txBuilder
                .validity({
                firstValid: testData.testnetBlock.blockNumber,
                maxDuration: baseCoin.SWEEP_TXN_DURATION,
            })
                .referenceBlock(testData.testnetBlock.hash)
                .sender({ address: testData.unsignedSweepUser.walletAddress });
            const tx = await txBuilder.build();
            const txJson = tx.toJson();
            should_1.default.deepEqual(txJson.sender, testData.unsignedSweepUser.walletAddress);
            should_1.default.deepEqual(txJson.blockNumber, testData.testnetBlock.blockNumber);
            should_1.default.deepEqual(txJson.referenceBlock, testData.testnetBlock.hash);
            should_1.default.deepEqual(txJson.genesisHash, resources_1.genesisHash);
            should_1.default.deepEqual(txJson.specVersion, resources_1.specVersion);
            should_1.default.deepEqual(txJson.nonce, 0);
            should_1.default.deepEqual(txJson.tip, 0);
            should_1.default.deepEqual(txJson.transactionVersion, resources_1.txVersion);
            should_1.default.deepEqual(txJson.eraPeriod, baseCoin.SWEEP_TXN_DURATION);
        });
    });
    describe('Token Enablement:', function () {
        it('should have correct token enablement config', function () {
            const config = baseCoin.getTokenEnablementConfig();
            should_1.default.exist(config);
            config.requiresTokenEnablement.should.equal(true);
            config.supportsMultipleTokenEnablements.should.equal(false);
        });
        it('should validate wallet type for token enablement', function () {
            const config = baseCoin.getTokenEnablementConfig();
            (() => config.validateWallet('custodial')).should.not.throw();
            (() => config.validateWallet('hot')).should.throw(/Token enablement for Polymesh \(polyx\) is only supported for custodial wallets/);
            (() => config.validateWallet('cold')).should.throw(/Token enablement for Polymesh \(polyx\) is only supported for custodial wallets/);
        });
    });
});
