// @substrate/txwrapper-substrate/lib/methods/balances/transferAll.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferAll = void 0;
const txwrapper_core_1 = require("@substrate/txwrapper-core");
/**
 * Transfer the entire transferable balance from the caller account.
 *
 * NOTE: This function only attempts to transfer _transferable_ balances. This means that
 * any locked, reserved, or existential deposits (when `keep_alive` is `true`), will not be
 * transferred by this function. To ensure that this function results in a killed account,
 * you might need to prepare the account by removing any reference counters, storage
 * deposits, etc...
 *
 * @param args - Arguments specific to this method.
 * @param info - Information required to construct the transaction.
 * @param options - Registry and metadata used for constructing the method.
 */
function transferAll(args, info, options) {
    return (0, txwrapper_core_1.defineMethod)({
        method: {
            args,
            name: 'transferAll',
            pallet: 'balances',
        },
        ...info,
    }, options);
}
exports.transferAll = transferAll;
