// @bitgo/sdk-coin-hash/dist/test/unit/transactionBuilder/contractCallBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const encoding_1 = require("@cosmjs/encoding");
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../src");
const hash_1 = require("../../resources/hash");
describe('Hash ContractCall Builder', () => {
    let bitgo;
    let basecoin;
    let factory;
    // Helper function to create a complete transaction builder with standard settings and message
    const contractCallBuilder = () => {
        const txBuilder = factory.getContractCallBuilder();
        txBuilder.sequence(hash_1.TEST_CONTRACT_CALL.sequence);
        txBuilder.accountNumber(hash_1.TEST_CONTRACT_CALL.accountNumber);
        txBuilder.chainId(hash_1.TEST_CONTRACT_CALL.chainId);
        txBuilder.gasBudget({
            amount: [{ denom: 'nhash', amount: hash_1.TEST_CONTRACT_CALL.fee }],
            gasLimit: hash_1.TEST_CONTRACT_CALL.gasLimit,
        });
        txBuilder.feeGranter(hash_1.TEST_CONTRACT_CALL.feeGranter);
        txBuilder.publicKey((0, encoding_1.toHex)((0, encoding_1.fromBase64)(hash_1.TEST_CONTRACT_CALL.pubKey)));
        txBuilder.messages([
            {
                sender: hash_1.TEST_CONTRACT_CALL.proposer,
                contract: hash_1.testnetAddress.groupPolicyAddress,
                msg: (0, encoding_1.fromBase64)(hash_1.TEST_CONTRACT_CALL.encodedProposal),
                funds: [],
            },
        ]);
        return txBuilder;
    };
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('hash', src_1.Hash.createInstance);
        bitgo.safeRegister('thash', src_1.Thash.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('thash');
        factory = basecoin.getBuilder();
    });
    // Helper function to create a group vote transaction builder
    const groupVoteBuilder = () => {
        const txBuilder = factory.getContractCallBuilder();
        txBuilder.sequence(hash_1.TEST_GROUP_VOTE.sequence);
        txBuilder.accountNumber(hash_1.TEST_GROUP_VOTE.accountNumber);
        txBuilder.chainId(hash_1.TEST_GROUP_VOTE.chainId);
        txBuilder.gasBudget({
            amount: [{ denom: 'nhash', amount: hash_1.TEST_GROUP_VOTE.fee }],
            gasLimit: hash_1.TEST_GROUP_VOTE.gasLimit,
        });
        txBuilder.feeGranter(hash_1.TEST_GROUP_VOTE.feeGranter);
        txBuilder.publicKey((0, encoding_1.toHex)((0, encoding_1.fromBase64)(hash_1.TEST_GROUP_VOTE.pubKey)));
        txBuilder.messages([
            {
                sender: hash_1.TEST_GROUP_VOTE.voter,
                contract: hash_1.testnetAddress.groupPolicyAddress,
                msg: (0, encoding_1.fromBase64)(hash_1.TEST_GROUP_VOTE.encodedVote),
                funds: [],
            },
        ]);
        return txBuilder;
    };
    describe('Contract Call Builder Tests', () => {
        it('should build transaction with expected signable payload', async function () {
            const txBuilder = contractCallBuilder();
            const tx = await txBuilder.build();
            should_1.default.equal((0, encoding_1.toHex)(tx.signablePayload), hash_1.TEST_CONTRACT_CALL.expectedSignBytesHex);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.ContractCall);
        });
        it('should build, sign, and serialize contract call transactions', async function () {
            // Test unsigned transaction building
            const unsignedBuilder = contractCallBuilder();
            const unsignedTx = await unsignedBuilder.build();
            should_1.default.equal(unsignedTx.type, sdk_core_1.TransactionType.ContractCall);
            should_1.default.equal(unsignedTx.signature.length, 0);
            should_1.default.exist(unsignedTx.toBroadcastFormat());
            // Test signing functionality
            const signedBuilder = contractCallBuilder();
            signedBuilder.sign({ key: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(hash_1.TEST_CONTRACT_CALL.privateKey)) });
            const signedTx = await signedBuilder.build();
            should_1.default.equal(signedTx.type, sdk_core_1.TransactionType.ContractCall);
            should_1.default.equal(signedTx.signature.length, 1);
            should_1.default.exist(signedTx.toBroadcastFormat());
        });
        it('should handle round-trip serialization for signed and unsigned transactions', async function () {
            // Test unsigned serialization
            const unsignedBuilder = contractCallBuilder();
            const unsignedTx = await unsignedBuilder.build();
            const unsignedRaw = unsignedTx.toBroadcastFormat();
            const rebuiltUnsigned = factory.from(unsignedRaw);
            const rebuiltUnsignedTx = await rebuiltUnsigned.build();
            should_1.default.equal(rebuiltUnsignedTx.toBroadcastFormat(), unsignedRaw);
            should_1.default.equal(rebuiltUnsignedTx.type, sdk_core_1.TransactionType.ContractCall);
            // Test signed serialization
            const signedBuilder = contractCallBuilder();
            signedBuilder.sign({ key: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(hash_1.TEST_CONTRACT_CALL.privateKey)) });
            const signedTx = await signedBuilder.build();
            const signedRaw = signedTx.toBroadcastFormat();
            const rebuiltSigned = factory.from(signedRaw);
            const rebuiltSignedTx = await rebuiltSigned.build();
            should_1.default.equal(rebuiltSignedTx.toBroadcastFormat(), signedRaw);
            should_1.default.equal(rebuiltSignedTx.signature.length, 1);
            should_1.default.equal(rebuiltSignedTx.signature[0], signedTx.signature[0]);
        });
    });
    describe('Group Vote Builder Tests', () => {
        it('should build vote transaction with expected signable payload', async function () {
            const txBuilder = groupVoteBuilder();
            const tx = await txBuilder.build();
            should_1.default.equal((0, encoding_1.toHex)(tx.signablePayload), hash_1.TEST_GROUP_VOTE.expectedSignBytesHex);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.ContractCall);
        });
        it('should build, sign, and serialize group vote transactions', async function () {
            const unsignedBuilder = groupVoteBuilder();
            const unsignedTx = await unsignedBuilder.build();
            should_1.default.equal(unsignedTx.type, sdk_core_1.TransactionType.ContractCall);
            should_1.default.equal(unsignedTx.signature.length, 0);
            should_1.default.exist(unsignedTx.toBroadcastFormat());
            const signedBuilder = groupVoteBuilder();
            signedBuilder.sign({ key: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(hash_1.TEST_GROUP_VOTE.privateKey)) });
            const signedTx = await signedBuilder.build();
            should_1.default.equal(signedTx.type, sdk_core_1.TransactionType.ContractCall);
            should_1.default.equal(signedTx.signature.length, 1);
            should_1.default.exist(signedTx.toBroadcastFormat());
        });
        it('should handle round-trip serialization for group vote transactions', async function () {
            const unsignedBuilder = groupVoteBuilder();
            const unsignedTx = await unsignedBuilder.build();
            const unsignedRaw = unsignedTx.toBroadcastFormat();
            const rebuiltUnsigned = factory.from(unsignedRaw);
            const rebuiltUnsignedTx = await rebuiltUnsigned.build();
            should_1.default.equal(rebuiltUnsignedTx.toBroadcastFormat(), unsignedRaw);
            should_1.default.equal(rebuiltUnsignedTx.type, sdk_core_1.TransactionType.ContractCall);
            const signedBuilder = groupVoteBuilder();
            signedBuilder.sign({ key: (0, encoding_1.toHex)((0, encoding_1.fromBase64)(hash_1.TEST_GROUP_VOTE.privateKey)) });
            const signedTx = await signedBuilder.build();
            const signedRaw = signedTx.toBroadcastFormat();
            const rebuiltSigned = factory.from(signedRaw);
            const rebuiltSignedTx = await rebuiltSigned.build();
            should_1.default.equal(rebuiltSignedTx.toBroadcastFormat(), signedRaw);
            should_1.default.equal(rebuiltSignedTx.signature.length, 1);
            should_1.default.equal(rebuiltSignedTx.signature[0], signedTx.signature[0]);
        });
    });
});
