// @bitgo/sdk-coin-xtz/dist/src/xtz.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Xtz = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const secp256k1_1 = require("@bitgo/secp256k1");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const lib_1 = require("./lib");
const utils_1 = require("./lib/utils");
const superagent_1 = __importDefault(require("superagent"));
class Xtz extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Xtz(bitgo, staticsCoin);
    }
    getChain() {
        return this._staticsCoin.name;
    }
    getFamily() {
        return this._staticsCoin.family;
    }
    getFullName() {
        return this._staticsCoin.fullName;
    }
    getBaseFactor() {
        return Math.pow(10, this._staticsCoin.decimalPlaces);
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.onchain;
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return true;
    }
    /**
     * Xtz supports transfers to consolidate balance from receive address to the wallet contract
     */
    allowsAccountConsolidations() {
        return true;
    }
    /**
     * Checks if this is a valid base58 or hex address
     * @param address
     */
    isValidAddress(address) {
        if (!address) {
            return false;
        }
        return lib_1.Utils.isValidAddress(address);
    }
    /**
     * Generate Tezos key pair - BitGo xpub format
     *
     * @param seed
     * @returns {Object} object with generated xpub, xprv
     */
    generateKeyPair(seed) {
        const keyPair = seed ? new lib_1.KeyPair({ seed }) : new lib_1.KeyPair();
        const keys = keyPair.getExtendedKeys();
        if (!keys.xprv) {
            throw new Error('Missing xprv in key generation.');
        }
        return {
            pub: keys.xpub,
            prv: keys.xprv,
        };
    }
    async parseTransaction(params) {
        return {};
    }
    async isWalletAddress(params) {
        throw new sdk_core_1.MethodNotImplementedError();
    }
    async verifyTransaction(params) {
        const { txParams } = params;
        if (Array.isArray(txParams.recipients) && txParams.recipients.length > 1) {
            throw new Error(`${this.getChain()} doesn't support sending to more than 1 destination address within a single transaction. Try again, using only a single recipient.`);
        }
        return true;
    }
    /**
     * Derive a user key using the chain path of the address
     * @param key
     * @param path
     * @returns {string} derived private key
     */
    deriveKeyWithPath({ key, path }) {
        const keychain = secp256k1_1.bip32.fromBase58(key);
        const derivedKeyNode = keychain.derivePath(path);
        return derivedKeyNode.toBase58();
    }
    /**
     * Assemble keychain and half-sign prebuilt transaction
     *
     * @param params
     * @param params.txPrebuild {Object} prebuild object returned by platform
     * @param params.prv {String} user prv
     * @returns Bluebird<SignedTransaction>
     */
    async signTransaction(params) {
        const txBuilder = new lib_1.TransactionBuilder(statics_1.coins.get(this.getChain()));
        txBuilder.from(params.txPrebuild.txHex);
        txBuilder.source(params.txPrebuild.source);
        if (params.txPrebuild.dataToSign) {
            txBuilder.overrideDataToSign({ dataToSign: params.txPrebuild.dataToSign });
        }
        // The path /0/0/0/0 is used by the wallet base address
        // Derive the user key only if the transaction is sent from a receive address
        let key;
        const { chain, index } = params.txPrebuild.addressInfo;
        if (chain === 0 && index === 0) {
            key = params.prv;
        }
        else {
            const derivationPath = `0/0/${chain}/${index}`;
            key = this.deriveKeyWithPath({ key: params.prv, path: derivationPath });
        }
        txBuilder.sign({ key });
        const transaction = await txBuilder.build();
        if (!transaction) {
            throw new Error('Invalid messaged passed to signMessage');
        }
        const response = {
            txHex: transaction.toBroadcastFormat(),
        };
        return transaction.signature.length >= 2 ? response : { halfSigned: response };
    }
    /**
     * Sign message with private key
     *
     * @param key
     * @param message
     */
    async signMessage(key, message) {
        const keyPair = new lib_1.KeyPair({ prv: key.prv });
        const messageHex = message instanceof Buffer ? message.toString('hex') : Buffer.from(message, 'utf-8').toString('hex');
        const signatureData = await lib_1.Utils.sign(keyPair, messageHex);
        return Buffer.from(signatureData.sig);
    }
    /**
     * Method to validate recovery params
     * @param {RecoverOptions} params
     * @returns {void}
     */
    validateRecoveryParams(params) {
        if (params.userKey === undefined) {
            throw new Error('missing userKey');
        }
        if (params.backupKey === undefined) {
            throw new Error('missing backupKey');
        }
        if (!params.isUnsignedSweep && params.walletPassphrase === undefined && !params.userKey.startsWith('xpub')) {
            throw new Error('missing wallet passphrase');
        }
        if (params.walletContractAddress === undefined || !this.isValidAddress(params.walletContractAddress)) {
            throw new Error('invalid walletContractAddress');
        }
        if (params.recoveryDestination === undefined || !this.isValidAddress(params.recoveryDestination)) {
            throw new Error('invalid recoveryDestination');
        }
    }
    /**
     * Make a query to blockchain explorer for information such as balance, token balance, solidity calls
     * @param query {Object} key-value pairs of parameters to append after /api
     * @param apiKey {string} optional API key to use instead of the one from the environment
     * @returns {Object} response from the blockchain explorer
     */
    async recoveryBlockchainExplorerQuery(params, apiKey) {
        const response = await superagent_1.default.get(`${sdk_core_1.common.Environments[this.bitgo.getEnv()].xtzExplorerBaseUrl}/v1/${params.actionPath}${params.address ? '/' + params.address : ''}${params.action ? '/' + params.action : ''}${apiKey ? `?apikey=${apiKey}` : ''}`);
        if (!response.ok) {
            throw new Error('could not reach TZKT');
        }
        if (response.status === 429) {
            throw new Error('TZKT rate limit reached');
        }
        return response.body;
    }
    /**
     * Queries public block explorer to get the next XTZ address details
     * @param {string} address
     * @param {string} apiKey - optional API key to use instead of the one from the environment
     * @returns {Promise<any>}
     */
    async getAddressDetails(address, apiKey) {
        const result = await this.recoveryBlockchainExplorerQuery({
            actionPath: 'accounts',
            address,
        }, apiKey);
        if (!result) {
            throw new Error(`Unable to find details for ${address}`);
        }
        return result;
    }
    /**
     * Query explorer for the balance of an address
     * @param {String} address - the XTZ base/receive address
     * @param {String} apiKey - optional API key to use instead of the one from the environment
     * @returns {BigNumber} address balance
     */
    async queryAddressBalance(address, apiKey) {
        const result = await this.recoveryBlockchainExplorerQuery({
            actionPath: (0, utils_1.isValidOriginatedAddress)(address) ? 'contracts' : 'accounts',
            address,
        }, apiKey);
        // throw if the result does not exist or the result is not a valid number
        if (!result || !result.balance) {
            throw new Error(`Could not obtain address balance for ${address} from the explorer`);
        }
        return new bignumber_js_1.default(result.balance, 10);
    }
    /**
     * Generate and pack the data to sign for each transfer.
     *
     * @param {String} contractAddress Wallet address to withdraw funds from
     * @param {String} contractCounter Wallet internal counter
     * @param {String} destination Tezos address to send the funds to
     * @param {String} amount Number of mutez to move
     * @param {IMSClient} imsClient Existing IMS client connection to reuse
     * @return {String} data to sign in hex format
     */
    async packDataToSign(contractAddress, contractCounter, destination, amount) {
        const dataToSign = (0, utils_1.generateDataToSign)(contractAddress, destination, amount, contractCounter);
        const xtzRpcUrl = `${sdk_core_1.common.Environments[this.bitgo.getEnv()].xtzRpcUrl}/chains/main/blocks/head/helpers/scripts/pack_data`;
        if (!xtzRpcUrl) {
            throw new Error('XTZ RPC url not found');
        }
        const response = await superagent_1.default.post(xtzRpcUrl).send(dataToSign);
        if (response.status === 404) {
            throw new Error(`unable to pack data to sign ${response.status}: ${response.body.error.message}`);
        }
        else if (response.status !== 200) {
            throw new Error(`unexpected IMS response status ${response.status}: ${response.body.error.message}`);
        }
        return response.body.packed;
    }
    /**
     * Builds a funds recovery transaction without BitGo.
     * We need to do three queries during this:
     * 1) Node query - how much money is in the account
     * 2) Build transaction - build our transaction for the amount
     * 3) Send signed build - send our signed build to a public node
     * @param params
     */
    async recover(params) {
        const isUnsignedSweep = params.isUnsignedSweep;
        this.validateRecoveryParams(params);
        // Clean up whitespace from entered values
        const backupKey = params.backupKey.replace(/\s/g, '');
        const userAddressDetails = await this.getAddressDetails(params.walletContractAddress, params.apiKey);
        if (!userAddressDetails) {
            throw new Error('Unable to fetch user address details');
        }
        // Decrypt backup private key and get address
        let backupPrv;
        let keyPair;
        let backupSigningKey;
        if (isUnsignedSweep) {
            keyPair = new lib_1.KeyPair({ pub: backupKey });
        }
        else {
            try {
                backupPrv = this.bitgo.decrypt({
                    input: backupKey,
                    password: params.walletPassphrase,
                });
            }
            catch (e) {
                throw new Error(`Error decrypting backup keychain: ${e.message}`);
            }
            keyPair = new lib_1.KeyPair({ prv: backupPrv });
            backupSigningKey = keyPair.getKeys().prv;
            if (!backupSigningKey) {
                throw new Error('no private key');
            }
        }
        const backupKeyAddress = keyPair.getAddress();
        const backupAddressDetails = await this.getAddressDetails(backupKeyAddress, params.apiKey || '');
        if (!backupAddressDetails.counter || !backupAddressDetails.balance) {
            throw new Error(`Missing required detail(s) for ${backupKeyAddress}: counter, balance`);
        }
        const backupKeyNonce = new bignumber_js_1.default(backupAddressDetails.counter + 1, 10);
        // get balance of backupKey to ensure funds are available to pay fees
        const backupKeyBalance = new bignumber_js_1.default(backupAddressDetails.balance, 10);
        const gasLimit = (0, utils_1.isValidOriginatedAddress)(params.recoveryDestination)
            ? utils_1.TRANSACTION_GAS_LIMIT.CONTRACT_TRANSFER
            : utils_1.TRANSACTION_GAS_LIMIT.TRANSFER;
        const gasPrice = utils_1.TRANSACTION_FEE.TRANSFER;
        // Checking whether back up key address has sufficient funds for transaction
        if (backupKeyBalance.lt(gasPrice)) {
            const weiToGwei = 10 ** 6;
            throw new Error(`Backup key address ${backupKeyAddress} has balance ${(backupKeyBalance.toNumber() / weiToGwei).toString()} Gwei.` +
                `This address must have a balance of at least ${(gasPrice / weiToGwei).toString()}` +
                ` Gwei to perform recoveries. Try sending some funds to this address then retry.`);
        }
        // get balance of sender address
        if (!userAddressDetails.balance || userAddressDetails.balance === 0) {
            throw new Error('No funds to recover from source address');
        }
        const txAmount = userAddressDetails.balance;
        if (new bignumber_js_1.default(txAmount).isLessThanOrEqualTo(0)) {
            throw new Error('Wallet does not have enough funds to recover');
        }
        const feeInfo = {
            fee: new bignumber_js_1.default(utils_1.TRANSACTION_FEE.TRANSFER),
            gasLimit: new bignumber_js_1.default(gasLimit),
            storageLimit: new bignumber_js_1.default(utils_1.TRANSACTION_STORAGE_LIMIT.TRANSFER),
        };
        const txBuilder = new lib_1.TransactionBuilder(statics_1.coins.get(this.getChain()));
        txBuilder.type(sdk_core_1.TransactionType.Send);
        txBuilder.source(backupKeyAddress);
        // Used to set the branch for the transaction
        const chainHead = await this.recoveryBlockchainExplorerQuery({
            actionPath: 'head',
        });
        if (!chainHead || !chainHead.hash) {
            throw new Error('Unable to fetch chain head');
        }
        txBuilder.branch(chainHead.hash);
        if (!backupAddressDetails.revealed) {
            feeInfo.fee = feeInfo.fee.plus(utils_1.TRANSACTION_FEE.REVEAL);
            feeInfo.gasLimit = feeInfo.gasLimit.plus(utils_1.TRANSACTION_GAS_LIMIT.REVEAL);
            feeInfo.storageLimit = feeInfo.storageLimit.plus(utils_1.TRANSACTION_STORAGE_LIMIT.REVEAL);
            backupKeyNonce.plus(1);
            const publicKeyToReveal = keyPair.getKeys();
            txBuilder.publicKeyToReveal(publicKeyToReveal.pub);
        }
        txBuilder.counter(backupKeyNonce.toString());
        const packedDataToSign = await this.packDataToSign(params.walletContractAddress, backupKeyNonce.toString(), params.recoveryDestination, txAmount?.toString());
        txBuilder
            .transfer(txAmount?.toString())
            .from(params.walletContractAddress)
            .to(params.recoveryDestination)
            .counter(backupKeyNonce.toString())
            .fee(utils_1.TRANSACTION_FEE.TRANSFER.toString())
            .storageLimit(utils_1.TRANSACTION_STORAGE_LIMIT.TRANSFER.toString())
            .gasLimit(gasLimit.toString())
            .dataToSign(packedDataToSign);
        if (isUnsignedSweep) {
            const tx = await txBuilder.build();
            const txInfo = tx.toJson();
            return {
                txHex: tx.toBroadcastFormat(),
                txInfo,
                source: params.walletContractAddress,
                dataToSign: packedDataToSign,
                feeInfo,
                sourceCounter: backupKeyNonce.toString(),
                transferCounters: [backupKeyNonce.toString()],
            };
        }
        txBuilder.sign({ key: backupSigningKey });
        const signedTx = await txBuilder.build();
        return {
            id: signedTx.id,
            tx: signedTx.toBroadcastFormat(),
        };
    }
    /**
     * Explain a Tezos transaction from txHex
     * @param params
     */
    async explainTransaction(params) {
        const txHex = params.txHex || (params.halfSigned && params.halfSigned.txHex);
        if (!txHex || !params.feeInfo) {
            throw new Error('missing explain tx parameters');
        }
        const txBuilder = new lib_1.TransactionBuilder(statics_1.coins.get(this.getChain()));
        // Newer coins can return BaseTransactionBuilderFactory instead of BaseTransactionBuilder
        if (!(txBuilder instanceof sdk_core_1.BaseTransactionBuilder)) {
            throw new Error('getBuilder() did not return an BaseTransactionBuilder object. Has it been updated?');
        }
        txBuilder.from(txHex);
        const tx = await txBuilder.build();
        const displayOrder = ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee'];
        return {
            displayOrder,
            id: tx.id,
            outputs: tx.outputs,
            outputAmount: tx.outputs
                .reduce((accumulator, output) => accumulator.plus(output.value), new bignumber_js_1.default('0'))
                .toFixed(0),
            changeOutputs: [], // account based does not use change outputs
            changeAmount: '0', // account base does not make change
            fee: params.feeInfo,
        };
    }
    isValidPub(pub) {
        return lib_1.Utils.isValidPublicKey(pub);
    }
    /** @inheritDoc */
    auditDecryptedKey(params) {
        throw new sdk_core_1.MethodNotImplementedError();
    }
}
exports.Xtz = Xtz;
