// @bitgo/wasm-utxo/dist/cjs/js/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasPsbtMagic = exports.ZcashTransaction = exports.Transaction = exports.DashTransaction = exports.Psbt = exports.Miniscript = exports.Descriptor = exports.isCoinName = exports.isTestnet = exports.isMainnet = exports.getMainnet = exports.Dimensions = exports.BIP32 = exports.ECPair = exports.ecpair = exports.bip32 = exports.descriptorWallet = exports.fixedScriptWallet = exports.utxolibCompat = exports.message = exports.inscriptions = exports.bip322 = exports.ast = exports.address = void 0;
const wasm = __importStar(require("./wasm/wasm_utxo.js"));
// we need to access the wasm module here, otherwise webpack gets all weird
// and forgets to include it in the bundle
void wasm;
// Most exports are namespaced to avoid polluting the top-level namespace
// and to make imports more explicit (e.g., `import { address } from '@bitgo/wasm-utxo'`)
exports.address = __importStar(require("./address.js"));
exports.ast = __importStar(require("./ast/index.js"));
exports.bip322 = __importStar(require("./bip322/index.js"));
exports.inscriptions = __importStar(require("./inscriptions.js"));
exports.message = __importStar(require("./message.js"));
exports.utxolibCompat = __importStar(require("./utxolibCompat.js"));
exports.fixedScriptWallet = __importStar(require("./fixedScriptWallet/index.js"));
exports.descriptorWallet = __importStar(require("./descriptorWallet/index.js"));
exports.bip32 = __importStar(require("./bip32.js"));
exports.ecpair = __importStar(require("./ecpair.js"));
// Only the most commonly used classes and types are exported at the top level for convenience
var ecpair_js_1 = require("./ecpair.js");
Object.defineProperty(exports, "ECPair", { enumerable: true, get: function () { return ecpair_js_1.ECPair; } });
var bip32_js_1 = require("./bip32.js");
Object.defineProperty(exports, "BIP32", { enumerable: true, get: function () { return bip32_js_1.BIP32; } });
var Dimensions_js_1 = require("./fixedScriptWallet/Dimensions.js");
Object.defineProperty(exports, "Dimensions", { enumerable: true, get: function () { return Dimensions_js_1.Dimensions; } });
var coinName_js_1 = require("./coinName.js");
Object.defineProperty(exports, "getMainnet", { enumerable: true, get: function () { return coinName_js_1.getMainnet; } });
Object.defineProperty(exports, "isMainnet", { enumerable: true, get: function () { return coinName_js_1.isMainnet; } });
Object.defineProperty(exports, "isTestnet", { enumerable: true, get: function () { return coinName_js_1.isTestnet; } });
Object.defineProperty(exports, "isCoinName", { enumerable: true, get: function () { return coinName_js_1.isCoinName; } });
var wasm_utxo_js_1 = require("./wasm/wasm_utxo.js");
Object.defineProperty(exports, "Descriptor", { enumerable: true, get: function () { return wasm_utxo_js_1.WrapDescriptor; } });
var wasm_utxo_js_2 = require("./wasm/wasm_utxo.js");
Object.defineProperty(exports, "Miniscript", { enumerable: true, get: function () { return wasm_utxo_js_2.WrapMiniscript; } });
var wasm_utxo_js_3 = require("./wasm/wasm_utxo.js");
Object.defineProperty(exports, "Psbt", { enumerable: true, get: function () { return wasm_utxo_js_3.WrapPsbt; } });
var transaction_js_1 = require("./transaction.js");
Object.defineProperty(exports, "DashTransaction", { enumerable: true, get: function () { return transaction_js_1.DashTransaction; } });
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return transaction_js_1.Transaction; } });
Object.defineProperty(exports, "ZcashTransaction", { enumerable: true, get: function () { return transaction_js_1.ZcashTransaction; } });
var psbt_js_1 = require("./psbt.js");
Object.defineProperty(exports, "hasPsbtMagic", { enumerable: true, get: function () { return psbt_js_1.hasPsbtMagic; } });
