// @bitgo/sdk-coin-sui/dist/src/sui.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sui = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const lib_1 = require("./lib");
const utils_1 = __importDefault(require("./lib/utils"));
const _ = __importStar(require("lodash"));
const iface_1 = require("./lib/iface");
const constants_1 = require("./lib/constants");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
class Sui extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Sui(bitgo, staticsCoin);
    }
    /**
     * Factor between the coin's base unit and its smallest subdivison
     */
    getBaseFactor() {
        return 1e9;
    }
    getChain() {
        return 'sui';
    }
    getFamily() {
        return 'sui';
    }
    getFullName() {
        return 'Sui';
    }
    getNetwork() {
        return this._staticsCoin.network;
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    getMPCAlgorithm() {
        return 'eddsa';
    }
    allowsAccountConsolidations() {
        return true;
    }
    async verifyTransaction(params) {
        let totalAmount = new bignumber_js_1.default(0);
        const coinConfig = statics_1.coins.get(this.getChain());
        const { txPrebuild: txPrebuild, txParams: txParams, verification, wallet } = params;
        const transaction = new lib_1.TransferTransaction(coinConfig);
        const rawTx = txPrebuild.txHex;
        if (!rawTx) {
            throw new Error('missing required tx prebuild property txHex');
        }
        transaction.fromRawTransaction(Buffer.from(rawTx, 'hex').toString('base64'));
        const explainedTx = transaction.explainTransaction();
        if (txParams.recipients && txParams.recipients.length > 0) {
            const filteredRecipients = txParams.recipients?.map((recipient) => {
                const filteredRecipient = _.pick(recipient, ['address', 'amount']);
                filteredRecipient.amount = new bignumber_js_1.default(filteredRecipient.amount).toFixed();
                return filteredRecipient;
            });
            const filteredOutputs = explainedTx.outputs.map((output) => {
                const filteredOutput = _.pick(output, ['address', 'amount']);
                filteredOutput.amount = new bignumber_js_1.default(filteredOutput.amount).toFixed();
                return filteredOutput;
            });
            if (!_.isEqual(filteredOutputs, filteredRecipients)) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            for (const recipients of txParams.recipients) {
                totalAmount = totalAmount.plus(recipients.amount);
            }
            if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                throw new Error('Tx total amount does not match with expected total amount field');
            }
        }
        if (verification?.consolidationToBaseAddress) {
            const baseAddress = wallet.coinSpecific()?.baseAddress || wallet.coinSpecific()?.rootAddress;
            for (const output of explainedTx.outputs) {
                if (output.address !== baseAddress) {
                    throw new Error('Consolidation transaction output address does not match wallet base address');
                }
            }
        }
        return true;
    }
    async isWalletAddress(params) {
        return (0, sdk_core_1.verifyEddsaTssWalletAddress)(params, (address) => this.isValidAddress(address), (publicKey) => this.getAddressFromPublicKey(publicKey));
    }
    async parseTransaction(params) {
        const transactionExplanation = await this.explainTransaction({ txHex: params.txHex });
        if (!transactionExplanation) {
            throw new Error('Invalid transaction');
        }
        let fee = new bignumber_js_1.default(0);
        const suiTransaction = transactionExplanation;
        if (suiTransaction.outputs.length <= 0) {
            return {
                inputs: [],
                outputs: [],
                fee,
            };
        }
        const senderAddress = suiTransaction.outputs[0].address;
        if (suiTransaction.fee.fee !== '') {
            fee = new bignumber_js_1.default(suiTransaction.fee.fee);
        }
        // assume 1 sender, who is also the fee payer
        const inputs = [
            {
                address: senderAddress,
                amount: new bignumber_js_1.default(suiTransaction.outputAmount).plus(fee).toFixed(),
            },
        ];
        const outputs = suiTransaction.outputs.map((output) => {
            return {
                address: output.address,
                amount: new bignumber_js_1.default(output.amount).toFixed(),
            };
        });
        return {
            inputs,
            outputs,
            fee,
        };
    }
    generateKeyPair(seed) {
        const keyPair = seed ? new lib_1.KeyPair({ seed }) : new lib_1.KeyPair();
        const keys = keyPair.getKeys();
        if (!keys.prv) {
            throw new Error('Missing prv in key generation.');
        }
        return {
            pub: keys.pub,
            prv: keys.prv,
        };
    }
    isValidPub(_) {
        throw new Error('Method not implemented.');
    }
    isValidPrv(_) {
        throw new Error('Method not implemented.');
    }
    isValidAddress(address) {
        return utils_1.default.isValidAddress(address);
    }
    signTransaction(_) {
        throw new Error('Method not implemented.');
    }
    /**
     * Explain a Sui transaction
     * @param params
     */
    async explainTransaction(params) {
        const factory = this.getBuilder();
        let rebuiltTransaction;
        try {
            const transactionBuilder = factory.from(Buffer.from(params.txHex, 'hex').toString('base64'));
            rebuiltTransaction = await transactionBuilder.build();
        }
        catch {
            throw new Error('Invalid transaction');
        }
        return rebuiltTransaction.explainTransaction();
    }
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    getAddressFromPublicKey(derivedPublicKey) {
        // TODO(BG-59016) replace with account lib implementation
        return utils_1.default.getAddressFromPublicKey(derivedPublicKey);
    }
    /** @inheritDoc */
    async getSignablePayload(serializedTx) {
        const factory = this.getBuilder();
        const rebuiltTransaction = await factory.from(serializedTx).build();
        return rebuiltTransaction.signablePayload;
    }
    getPublicNodeUrl() {
        return sdk_core_1.Environments[this.bitgo.getEnv()].suiNodeUrl;
    }
    async getBalance(owner, coinType) {
        const url = this.getPublicNodeUrl();
        return await utils_1.default.getBalance(url, owner, coinType);
    }
    async getInputCoins(owner, coinType) {
        const url = this.getPublicNodeUrl();
        return await utils_1.default.getInputCoins(url, owner, coinType);
    }
    async getFeeEstimate(txHex) {
        const url = this.getPublicNodeUrl();
        return await utils_1.default.getFeeEstimate(url, txHex);
    }
    /**
     * Builds funds recovery transaction(s) without BitGo
     *
     * @param {MPCRecoveryOptions} params parameters needed to construct and
     * (maybe) sign the transaction
     *
     * @returns {MPCTx | MPCSweepTxs} array of the serialized transaction hex strings and indices
     * of the addresses being swept
     */
    async recover(params) {
        if (!params.bitgoKey) {
            throw new Error('missing bitgoKey');
        }
        if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
            throw new Error('invalid recoveryDestination');
        }
        const startIdx = utils_1.default.validateNonNegativeNumber(0, 'Invalid starting index to scan for addresses', params.startingScanIndex);
        const numIterations = utils_1.default.validateNonNegativeNumber(20, 'Invalid scanning factor', params.scan);
        const endIdx = startIdx + numIterations;
        const bitgoKey = params.bitgoKey.replace(/\s/g, '');
        const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
        for (let idx = startIdx; idx < endIdx; idx++) {
            const derivationPath = (params.seed ? (0, sdk_lib_mpc_1.getDerivationPath)(params.seed) : 'm') + `/${idx}`;
            const derivedPublicKey = MPC.deriveUnhardened(bitgoKey, derivationPath).slice(0, 64);
            const senderAddress = this.getAddressFromPublicKey(derivedPublicKey);
            let availableBalance = new bignumber_js_1.default(0);
            try {
                availableBalance = new bignumber_js_1.default(await this.getBalance(senderAddress));
            }
            catch (e) {
                continue;
            }
            if (availableBalance.minus(constants_1.MAX_GAS_BUDGET).toNumber() <= 0) {
                continue;
            }
            // check for possible token recovery, recover the token provide by user
            if (params.tokenContractAddress) {
                const token = utils_1.default.getSuiTokenFromAddress(params.tokenContractAddress, this.getNetwork());
                if (!token) {
                    throw new Error(`Sui Token Package ID not supported.`);
                }
                const coinType = `${token.packageId}::${token.module}::${token.symbol}`;
                try {
                    const availableTokenBalance = new bignumber_js_1.default(await this.getBalance(senderAddress, coinType));
                    if (availableTokenBalance.toNumber() <= 0) {
                        continue;
                    }
                }
                catch (e) {
                    continue;
                }
                return this.recoverSuiToken(params, token, senderAddress, derivationPath, derivedPublicKey, idx, bitgoKey);
            }
            let inputCoins = await this.getInputCoins(senderAddress);
            inputCoins = inputCoins.sort((a, b) => {
                return b.balance.minus(a.balance).toNumber();
            });
            if (inputCoins.length > constants_1.MAX_OBJECT_LIMIT) {
                inputCoins = inputCoins.slice(0, constants_1.MAX_OBJECT_LIMIT);
            }
            let netAmount = inputCoins.reduce((acc, obj) => acc.plus(obj.balance), new bignumber_js_1.default(0));
            netAmount = netAmount.minus(constants_1.MAX_GAS_BUDGET);
            const recipients = [
                {
                    address: params.recoveryDestination,
                    amount: netAmount.toString(),
                },
            ];
            // first build the unsigned txn
            const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
            const txBuilder = factory
                .getTransferBuilder()
                .type(iface_1.SuiTransactionType.Transfer)
                .sender(senderAddress)
                .send(recipients)
                .gasData({
                owner: senderAddress,
                price: constants_1.DEFAULT_GAS_PRICE,
                budget: constants_1.MAX_GAS_BUDGET,
                payment: inputCoins,
            });
            const tempTx = (await txBuilder.build());
            const feeEstimate = await this.getFeeEstimate(tempTx.toBroadcastFormat());
            const gasBudget = Math.trunc(feeEstimate.toNumber() * constants_1.DEFAULT_GAS_OVERHEAD);
            netAmount = netAmount.plus(constants_1.MAX_GAS_BUDGET).minus(gasBudget);
            recipients[0].amount = netAmount.toString();
            txBuilder.send(recipients);
            txBuilder.gasData({
                owner: senderAddress,
                price: constants_1.DEFAULT_GAS_PRICE,
                budget: gasBudget,
                payment: inputCoins,
            });
            const isUnsignedSweep = !params.walletPassphrase;
            if (isUnsignedSweep) {
                return this.buildUnsignedSweepTransaction(txBuilder, senderAddress, bitgoKey, idx, derivationPath);
            }
            await this.signRecoveryTransaction(txBuilder, params, derivationPath, derivedPublicKey, false);
            const tx = (await txBuilder.build());
            return {
                transactions: [
                    {
                        scanIndex: idx,
                        recoveryAmount: netAmount.toString(),
                        serializedTx: tx.toBroadcastFormat(),
                        signature: Buffer.from(tx.serializedSig).toString('base64'),
                        coin: this.getChain(),
                    },
                ],
                lastScanIndex: idx,
            };
        }
        throw new Error(`Did not find an address with sufficient funds to recover. Please start the next scan at address index ${endIdx}. If it is token transaction, please keep sufficient Sui balance in the address for the transaction fee.`);
    }
    async recoverSuiToken(params, token, senderAddress, derivationPath, derivedPublicKey, idx, bitgoKey) {
        const coinType = `${token.packageId}::${token.module}::${token.symbol}`;
        let tokenObjects = await this.getInputCoins(senderAddress, coinType);
        tokenObjects = tokenObjects.sort((a, b) => {
            return b.balance.minus(a.balance).toNumber();
        });
        if (tokenObjects.length > constants_1.TOKEN_OBJECT_LIMIT) {
            tokenObjects = tokenObjects.slice(0, constants_1.TOKEN_OBJECT_LIMIT);
        }
        const netAmount = tokenObjects.reduce((acc, obj) => acc.plus(obj.balance), new bignumber_js_1.default(0));
        const recipients = [
            {
                address: params.recoveryDestination,
                amount: netAmount.toString(),
            },
        ];
        const gasAmount = new bignumber_js_1.default(constants_1.MAX_GAS_BUDGET);
        let gasObjects = await this.getInputCoins(senderAddress);
        gasObjects = utils_1.default.selectObjectsInDescOrderOfBalance(gasObjects, gasAmount);
        if (gasObjects.length >= constants_1.MAX_GAS_OBJECTS) {
            gasObjects = gasObjects.slice(0, constants_1.MAX_GAS_OBJECTS - 1);
        }
        // first build the unsigned txn
        const factory = new lib_1.TransactionBuilderFactory(token);
        const txBuilder = factory
            .getTokenTransferBuilder()
            .type(iface_1.SuiTransactionType.TokenTransfer)
            .sender(senderAddress)
            .send(recipients)
            .inputObjects(tokenObjects)
            .gasData({
            owner: senderAddress,
            price: constants_1.DEFAULT_GAS_PRICE,
            budget: constants_1.MAX_GAS_BUDGET,
            payment: gasObjects,
        });
        const tempTx = (await txBuilder.build());
        const feeEstimate = await this.getFeeEstimate(tempTx.toBroadcastFormat());
        const gasBudget = Math.trunc(feeEstimate.toNumber() * constants_1.DEFAULT_GAS_OVERHEAD);
        txBuilder.gasData({
            owner: senderAddress,
            price: constants_1.DEFAULT_GAS_PRICE,
            budget: gasBudget,
            payment: gasObjects,
        });
        const isUnsignedSweep = !params.walletPassphrase;
        if (isUnsignedSweep) {
            return this.buildUnsignedSweepTransaction(txBuilder, senderAddress, bitgoKey, idx, derivationPath, token);
        }
        await this.signRecoveryTransaction(txBuilder, params, derivationPath, derivedPublicKey, true);
        const tx = (await txBuilder.build());
        return {
            transactions: [
                {
                    scanIndex: idx,
                    recoveryAmount: netAmount.toString(),
                    serializedTx: tx.toBroadcastFormat(),
                    signature: Buffer.from(tx.serializedSig).toString('base64'),
                    coin: token.name,
                },
            ],
            lastScanIndex: idx,
        };
    }
    async buildUnsignedSweepTransaction(txBuilder, senderAddress, bitgoKey, index, derivationPath, token) {
        const isTokenTransaction = !!token;
        const unsignedTransaction = isTokenTransaction
            ? (await txBuilder.build())
            : (await txBuilder.build());
        const serializedTx = unsignedTransaction.toBroadcastFormat();
        const serializedTxHex = Buffer.from(serializedTx, 'base64').toString('hex');
        const parsedTx = await this.parseTransaction({ txHex: serializedTxHex });
        const walletCoin = isTokenTransaction ? token.name : this.getChain();
        const output = parsedTx.outputs[0];
        const inputs = [
            {
                address: senderAddress,
                valueString: output.amount,
                value: new bignumber_js_1.default(output.amount),
            },
        ];
        const outputs = [
            {
                address: output.address,
                valueString: output.amount,
                coinName: walletCoin,
            },
        ];
        const spendAmount = output.amount;
        const completedParsedTx = {
            inputs: inputs,
            outputs: outputs,
            spendAmount: spendAmount,
            type: isTokenTransaction ? iface_1.SuiTransactionType.TokenTransfer : iface_1.SuiTransactionType.Transfer,
        };
        const fee = parsedTx.fee;
        const feeInfo = { fee: fee.toNumber(), feeString: fee.toString() };
        const coinSpecific = { commonKeychain: bitgoKey };
        const transaction = {
            serializedTx: serializedTxHex,
            scanIndex: index,
            coin: walletCoin,
            signableHex: unsignedTransaction.signablePayload.toString('hex'),
            derivationPath,
            parsedTx: completedParsedTx,
            feeInfo: feeInfo,
            coinSpecific: coinSpecific,
        };
        const unsignedTx = { unsignedTx: transaction, signatureShares: [] };
        const transactions = [unsignedTx];
        const txRequest = {
            transactions: transactions,
            walletCoin: walletCoin,
        };
        return { txRequests: [txRequest] };
    }
    async signRecoveryTransaction(txBuilder, params, derivationPath, derivedPublicKey, isTokenTransaction) {
        // TODO(BG-51092): This looks like a common part which can be extracted out too
        const unsignedTx = isTokenTransaction
            ? (await txBuilder.build())
            : (await txBuilder.build());
        if (!params.userKey) {
            throw new Error('missing userKey');
        }
        if (!params.backupKey) {
            throw new Error('missing backupKey');
        }
        if (!params.walletPassphrase) {
            throw new Error('missing wallet passphrase');
        }
        // Clean up whitespace from entered values
        const userKey = params.userKey.replace(/\s/g, '');
        const backupKey = params.backupKey.replace(/\s/g, '');
        // Decrypt private keys from KeyCard values
        let userPrv;
        try {
            userPrv = this.bitgo.decrypt({
                input: userKey,
                password: params.walletPassphrase,
            });
        }
        catch (e) {
            throw new Error(`Error decrypting user keychain: ${e.message}`);
        }
        /** TODO BG-52419 Implement Codec for parsing */
        const userSigningMaterial = JSON.parse(userPrv);
        let backupPrv;
        try {
            backupPrv = this.bitgo.decrypt({
                input: backupKey,
                password: params.walletPassphrase,
            });
        }
        catch (e) {
            throw new Error(`Error decrypting backup keychain: ${e.message}`);
        }
        const backupSigningMaterial = JSON.parse(backupPrv);
        /* ********************** END ***********************************/
        // add signature
        const signatureHex = await sdk_core_1.EDDSAMethods.getTSSSignature(userSigningMaterial, backupSigningMaterial, derivationPath, unsignedTx);
        txBuilder.addSignature({ pub: derivedPublicKey }, signatureHex);
    }
    async broadcastTransaction({ transactions, }) {
        const txIds = [];
        const url = this.getPublicNodeUrl();
        let digest = '';
        if (!!transactions) {
            for (const txn of transactions) {
                try {
                    digest = await utils_1.default.executeTransactionBlock(url, txn.serializedTx, [txn.signature]);
                }
                catch (e) {
                    throw new Error(`Failed to broadcast transaction, error: ${e.message}`);
                }
                txIds.push(digest);
            }
        }
        return { txIds };
    }
    /** inherited doc */
    async createBroadcastableSweepTransaction(params) {
        const req = params.signatureShares;
        const broadcastableTransactions = [];
        let lastScanIndex = 0;
        for (let i = 0; i < req.length; i++) {
            const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
            const transaction = req[i].txRequest.transactions[0].unsignedTx;
            if (!req[i].ovc || !req[i].ovc[0].eddsaSignature) {
                throw new Error('Missing signature(s)');
            }
            const signature = req[i].ovc[0].eddsaSignature;
            if (!transaction.signableHex) {
                throw new Error('Missing signable hex');
            }
            const messageBuffer = Buffer.from(transaction.signableHex, 'hex');
            const result = MPC.verify(messageBuffer, signature);
            if (!result) {
                throw new Error('Invalid signature');
            }
            const signatureHex = Buffer.concat([Buffer.from(signature.R, 'hex'), Buffer.from(signature.sigma, 'hex')]);
            const serializedTxBase64 = Buffer.from(transaction.serializedTx, 'hex').toString('base64');
            const txBuilder = this.getBuilder().from(serializedTxBase64);
            if (!transaction.coinSpecific?.commonKeychain) {
                throw new Error('Missing common keychain');
            }
            const commonKeychain = transaction.coinSpecific.commonKeychain;
            if (!transaction.derivationPath) {
                throw new Error('Missing derivation path');
            }
            const derivationPath = transaction.derivationPath;
            const derivedPublicKey = MPC.deriveUnhardened(commonKeychain, derivationPath).slice(0, 64);
            // add combined signature from ovc
            txBuilder.addSignature({ pub: derivedPublicKey }, signatureHex);
            const signedTransaction = (await txBuilder.build());
            const serializedTx = signedTransaction.toBroadcastFormat();
            const outputAmount = signedTransaction.explainTransaction().outputAmount;
            broadcastableTransactions.push({
                serializedTx: serializedTx,
                scanIndex: transaction.scanIndex,
                signature: Buffer.from(signedTransaction.serializedSig).toString('base64'),
                recoveryAmount: outputAmount.toString(),
            });
            if (i === req.length - 1 && transaction.coinSpecific.lastScanIndex) {
                lastScanIndex = transaction.coinSpecific.lastScanIndex;
            }
        }
        return { transactions: broadcastableTransactions, lastScanIndex };
    }
    /**
     * Builds native SUI recoveries of receive addresses in batch without BitGo.
     * Funds will be recovered to base address first. You need to initiate another sweep txn after that.
     *
     * @param {MPCConsolidationRecoveryOptions} params - options for consolidation recovery.
     * @param {string} [params.startingScanIndex] - receive address index to start scanning from. default to 1 (inclusive).
     * @param {string} [params.endingScanIndex] - receive address index to end scanning at. default to startingScanIndex + 20 (exclusive).
     */
    async recoverConsolidations(params) {
        const isUnsignedSweep = !params.walletPassphrase;
        const startIdx = utils_1.default.validateNonNegativeNumber(1, 'Invalid starting index to scan for addresses', params.startingScanIndex);
        const endIdx = utils_1.default.validateNonNegativeNumber(startIdx + constants_1.DEFAULT_SCAN_FACTOR, 'Invalid ending index to scan for addresses', params.endingScanIndex);
        if (startIdx < 1 || endIdx <= startIdx || endIdx - startIdx > 10 * constants_1.DEFAULT_SCAN_FACTOR) {
            throw new Error(`Invalid starting or ending index to scan for addresses. startingScanIndex: ${startIdx}, endingScanIndex: ${endIdx}.`);
        }
        const bitgoKey = params.bitgoKey.replace(/\s/g, '');
        const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
        const derivationPath = (params.seed ? (0, sdk_lib_mpc_1.getDerivationPath)(params.seed) : 'm') + '/0';
        const derivedPublicKey = MPC.deriveUnhardened(bitgoKey, derivationPath).slice(0, 64);
        const baseAddress = this.getAddressFromPublicKey(derivedPublicKey);
        const consolidationTransactions = [];
        let lastScanIndex = startIdx;
        for (let idx = startIdx; idx < endIdx; idx++) {
            const recoverParams = {
                userKey: params.userKey,
                backupKey: params.backupKey,
                bitgoKey: params.bitgoKey,
                walletPassphrase: params.walletPassphrase,
                seed: params.seed,
                tokenContractAddress: params.tokenContractAddress,
                recoveryDestination: baseAddress,
                startingScanIndex: idx,
                scan: 1,
            };
            let recoveryTransaction;
            try {
                recoveryTransaction = await this.recover(recoverParams);
            }
            catch (e) {
                if (e.message.startsWith('Did not find an address with sufficient funds to recover.')) {
                    lastScanIndex = idx;
                    continue;
                }
                throw e;
            }
            if (isUnsignedSweep) {
                consolidationTransactions.push(recoveryTransaction.txRequests[0]);
            }
            else {
                consolidationTransactions.push(recoveryTransaction.transactions[0]);
            }
            lastScanIndex = idx;
        }
        if (consolidationTransactions.length === 0) {
            throw new Error(`Did not find an address with sufficient funds to recover. Please start the next scan at address index ${lastScanIndex + 1}.`);
        }
        if (isUnsignedSweep) {
            // lastScanIndex will be used to inform user the last address index scanned for available funds (so they can
            // appropriately adjust the scan range on the next iteration of consolidation recoveries). In the case of unsigned
            // sweep consolidations, this lastScanIndex will be provided in the coinSpecific of the last txn made.
            consolidationTransactions[consolidationTransactions.length - 1].transactions[0].unsignedTx.coinSpecific.lastScanIndex = lastScanIndex;
            return { txRequests: consolidationTransactions };
        }
        return { transactions: consolidationTransactions, lastScanIndex };
    }
    /** inherited doc */
    setCoinSpecificFieldsInIntent(intent, params) {
        intent.unspents = params.unspents;
    }
    /** inherited doc */
    auditDecryptedKey({ publicKey, prv, multiSigType }) {
        if (multiSigType !== 'tss') {
            throw new Error('Unsupported multiSigType');
        }
        (0, sdk_lib_mpc_1.auditEddsaPrivateKey)(prv, publicKey ?? '');
    }
}
exports.Sui = Sui;
