// bitgo/dist/test/v2/unit/enterprise.js
"use strict";
//
// Tests for enterprises
//
Object.defineProperty(exports, "__esModule", { value: true });
const nock = require("nock");
const sinon = require("sinon");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const bitgo_1 = require("../../../src/bitgo");
const ecdsaNtilde_1 = require("./internal/tssUtils/mocks/ecdsaNtilde");
const utxo_lib_1 = require("@bitgo/utxo-lib");
describe('Enterprise:', function () {
    let bitgo;
    let enterprise;
    let baseCoin;
    let bgUrl;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(bitgo_1.BitGo, { env: 'test' });
        bitgo.initializeTestVars();
        baseCoin = bitgo.coin('tbtc');
        enterprise = new sdk_core_1.Enterprise(bitgo, baseCoin, { id: '593f1ece99d37c23080a557283edcc89', name: 'Test Enterprise' });
        bgUrl = sdk_core_1.common.Environments[bitgo.getEnv()].uri;
    });
    describe('Transaction data', function () {
        it('should search for pending transaction correctly', async function () {
            const params = { enterpriseId: enterprise.id };
            const scope = nock(bgUrl).get('/api/v2/tbtc/tx/pending/first').query(params).reply(200);
            await enterprise.getFirstPendingTransaction().should.be.resolved();
            scope.isDone().should.be.True();
        });
    });
    it('should fetch the tss config correctly', async function () {
        const scope = nock(bgUrl)
            .get(`/api/v2/enterprise/${enterprise.id}/tssconfig`)
            .reply(200, {
            ecdsa: {
                challenge: {
                    enterprise: {
                        ...ecdsaNtilde_1.mockChallengeA,
                        verifiers: {
                            adminSignature: 'hex sig',
                        },
                    },
                },
            },
        });
        await enterprise.getExistingTssEcdsaChallenge().should.be.resolved();
        scope.isDone().should.be.True();
    });
    describe('Resign enterprise challenge', function () {
        const bitgo = sdk_test_1.TestBitGo.decorate(bitgo_1.BitGo, { env: 'mock' });
        const enterprise = new sdk_core_1.Enterprise(bitgo, bitgo.coin('tbtc'), { id: '123', name: 'Test Enterprise' });
        const oldAdminEcdhKey = bitgo.keychains().create();
        const oldAdminDerivationPath = 'm/0/0';
        const newAdminEcdhKey = bitgo.keychains().create();
        const newAdminDerivationPath = 'm/0/1';
        const entChallenge = {
            ntilde: 'ent ntilde',
            h1: 'ent h1',
            h2: 'ent h2',
        };
        const bitgoChallenge = {
            ntilde: 'bitgo ntilde',
            h1: 'bitgo h1',
            h2: 'bitgo h2',
        };
        const signedEntChallenge = sdk_core_1.ECDSAUtils.EcdsaUtils.signChallenge(entChallenge, oldAdminEcdhKey.xprv, oldAdminDerivationPath);
        const signedBitgoChallenge = sdk_core_1.ECDSAUtils.EcdsaUtils.signChallenge(bitgoChallenge, oldAdminEcdhKey.xprv, oldAdminDerivationPath);
        const newSignedEntChallenge = sdk_core_1.ECDSAUtils.EcdsaUtils.signChallenge(entChallenge, newAdminEcdhKey.xprv, newAdminDerivationPath);
        const newSignedBitgoChallenge = sdk_core_1.ECDSAUtils.EcdsaUtils.signChallenge(bitgoChallenge, newAdminEcdhKey.xprv, newAdminDerivationPath);
        const entChallengeWithVerifiers = {
            ...entChallenge,
            verifiers: {
                adminSignature: signedEntChallenge.toString('hex'),
            },
        };
        const bitgoChallengeWithVerifier = {
            ...bitgoChallenge,
            verifiers: {
                adminSignature: signedBitgoChallenge.toString('hex'),
            },
        };
        it('should verify and resign enterprise challenge', async function () {
            const stubuUploadChallenges = sinon.stub(sdk_core_1.ECDSAUtils.EcdsaUtils, 'uploadChallengesToEnterprise');
            await enterprise
                .resignEnterpriseChallenges({
                xprv: oldAdminEcdhKey.xprv,
                derivationPath: oldAdminDerivationPath,
                derivedPubKey: utxo_lib_1.bip32
                    .fromBase58(oldAdminEcdhKey.xpub)
                    .derivePath(oldAdminDerivationPath)
                    .publicKey.toString('hex'),
            }, {
                xprv: newAdminEcdhKey.xprv,
                derivationPath: newAdminDerivationPath,
                derivedPubKey: utxo_lib_1.bip32
                    .fromBase58(newAdminEcdhKey.xpub)
                    .derivePath(newAdminDerivationPath)
                    .publicKey.toString('hex'),
            }, entChallengeWithVerifiers, bitgoChallengeWithVerifier, bitgoChallengeWithVerifier)
                .should.not.be.rejected();
            stubuUploadChallenges.should.be.calledWith(bitgo, '123', entChallengeWithVerifiers, newSignedEntChallenge.toString('hex'), newSignedBitgoChallenge.toString('hex'), newSignedBitgoChallenge.toString('hex'));
        });
        it('should fail when the old ecdh keychain is incorrect', async function () {
            await enterprise
                .resignEnterpriseChallenges({
                xprv: newAdminEcdhKey.xprv,
                derivationPath: newAdminDerivationPath,
                derivedPubKey: utxo_lib_1.bip32
                    .fromBase58(newAdminEcdhKey.xpub)
                    .derivePath(newAdminDerivationPath)
                    .publicKey.toString('hex'),
            }, {
                xprv: newAdminEcdhKey.xprv,
                derivationPath: newAdminDerivationPath,
                derivedPubKey: utxo_lib_1.bip32
                    .fromBase58(newAdminEcdhKey.xpub)
                    .derivePath(newAdminDerivationPath)
                    .publicKey.toString('hex'),
            }, entChallengeWithVerifiers, bitgoChallengeWithVerifier, bitgoChallengeWithVerifier)
                .should.be.rejectedWith('Cannot re-sign. The Enterprise TSS config was signed by another user.');
        });
    });
});
