// bitgo/dist/test/v2/unit/pendingApproval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @prettier
 */
require("should");
const nock = require("nock");
const sinon = require("sinon");
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../../src");
const sdk_core_1 = require("@bitgo/sdk-core");
describe('Pending Approvals:', () => {
    let sandbox;
    let bitgo;
    let basecoin;
    let wallet;
    let bgUrl;
    const coin = 'tbtc';
    const walletId = 'wallet_id';
    const pendingApprovalData = {
        id: 'pa0',
        info: {
            type: sdk_core_1.Type.TRANSACTION_REQUEST,
            transactionRequest: {
                coinSpecific: {
                    [coin]: {},
                },
                recipients: [],
                buildParams: {
                    type: 'consolidate',
                },
                sourceWallet: walletId,
            },
        },
        state: sdk_core_1.State.PENDING,
        creator: 'test',
    };
    const walletData = {
        id: walletId,
        coin,
        pendingApprovals: [pendingApprovalData],
    };
    beforeEach(function () {
        sandbox = sinon.createSandbox();
        nock.disableNetConnect();
    });
    afterEach(function () {
        sandbox.restore();
    });
    before(async () => {
        // create wallet
        bitgo = sdk_test_1.TestBitGo.decorate(src_1.BitGo, { env: 'mock' });
        bitgo.initializeTestVars();
        basecoin = bitgo.coin(coin);
        wallet = new sdk_core_1.Wallet(bitgo, basecoin, walletData);
        bgUrl = sdk_core_1.Environments[bitgo.getEnv()].uri;
        pendingApprovalData.wallet = wallet;
    });
    ['tsol', 'teth', 'tbtc'].forEach((coinName) => {
        it(`should use correct tssUtils for  ${coinName}`, () => {
            const coin = bitgo.coin(coinName);
            const pendingAproval = new sdk_core_1.PendingApproval(bitgo, coin, {});
            if (coin.supportsTss()) {
                if (coin.getMPCAlgorithm() === 'ecdsa') {
                    pendingAproval['tssUtils'].should.be.instanceOf(sdk_core_1.ECDSAUtils.EcdsaUtils);
                }
                else if (coin.getMPCAlgorithm() === 'eddsa') {
                    pendingAproval['tssUtils'].should.be.instanceOf(sdk_core_1.EddsaUtils);
                }
            }
            else {
                (pendingAproval['tssUtils'] === undefined).should.be.true();
            }
        });
    });
    ['MPCv2', undefined].forEach((multisigTypeVersion) => {
        it(`should use correct tssUtils for multisigTypeVersion: ${multisigTypeVersion}`, () => {
            const coin = bitgo.coin('hteth');
            const walletDataMpcV2 = {
                ...walletData,
                multisigTypeVersion: multisigTypeVersion,
            };
            const walletMPCv2 = new sdk_core_1.Wallet(bitgo, basecoin, walletDataMpcV2);
            const pendingAproval = new sdk_core_1.PendingApproval(bitgo, coin, {}, walletMPCv2);
            if (walletMPCv2.multisigTypeVersion() === 'MPCv2') {
                pendingAproval['tssUtils'].should.be.instanceOf(sdk_core_1.ECDSAUtils.EcdsaMPCv2Utils);
            }
            else {
                pendingAproval['tssUtils'].should.be.instanceOf(sdk_core_1.ECDSAUtils.EcdsaUtils);
            }
        });
    });
    it('should call consolidate instead of build when rebuilding consolidation pending approvals', async () => {
        const scope = nock(bgUrl).post(`/api/v2/${coin}/wallet/${walletId}/consolidateUnspents`).reply(200);
        const pendingApprovals = wallet.pendingApprovals();
        pendingApprovals.should.have.length(1);
        const pendingApproval = pendingApprovals[0];
        // approval will fail when attempting to resign. This is ok - we just want to make sure
        // the consolidateUnspents endpoint was called already before failing
        await pendingApproval.approve({ xprv: 'nonsense' }).should.be.rejected();
        scope.done();
    });
    it('should approve for transactionRequestLite if we cannot recreate transaction', async () => {
        const pendingApprovalData2 = { ...pendingApprovalData, txRequestId: '1234-4567-6789' };
        const pendingApproval = new sdk_core_1.PendingApproval(bitgo, basecoin, pendingApprovalData2, wallet);
        const paScope = nock(bgUrl)
            .put(`/api/v2/${coin}/pendingapprovals/${pendingApprovalData.id}`, {
            state: 'approved',
            otp: undefined,
        })
            .reply(200, {
            ...pendingApprovalData2,
            state: 'approved',
        });
        const recreateTransactionTssStub = sandbox.stub(sdk_core_1.PendingApproval.prototype, 'recreateAndSignTSSTransaction');
        const recreateTransactionStub = sandbox.stub(sdk_core_1.PendingApproval.prototype, 'recreateAndSignTransaction');
        pendingApproval.type().should.equal(sdk_core_1.Type.TRANSACTION_REQUEST);
        await pendingApproval.approve({});
        recreateTransactionTssStub.notCalled.should.be.true();
        recreateTransactionStub.notCalled.should.be.true();
        paScope.isDone().should.be.true();
    });
    function testRecreateTransaction(coinName, recreateTransaction, type) {
        it(`[${coinName}] should ${recreateTransaction ? 'not ' : ''}recreate the transaction during approving a pending approval if there are no recipients for PA type ${type}`, async () => {
            const coin = bitgo.coin(coinName);
            const txRequestId = coin.supportsTss() ? 'requestTxIdTest' : undefined;
            const pendingApprovalInfo = type === 'transactionRequest'
                ? {
                    type,
                    transactionRequest: {
                        coinSpecific: {
                            [coinName]: { txHex: 'gabagool' },
                        },
                        recipients: [],
                        buildParams: {},
                        sourceWallet: walletId,
                    },
                }
                : { type };
            const pendingApprovalDataTemp = {
                id: 'pa0',
                info: pendingApprovalInfo,
                wallet: walletId,
                state: sdk_core_1.State.PENDING,
                creator: 'test',
                txRequestId,
            };
            const walletDataTemp = {
                id: walletId,
                coinName,
                pendingApprovals: [pendingApprovalDataTemp],
            };
            const walletTemp = new sdk_core_1.Wallet(bitgo, coin, walletDataTemp);
            pendingApprovalDataTemp.wallet = walletTemp.id();
            const pendingApprovals = walletTemp.pendingApprovals();
            pendingApprovals.should.have.length(1);
            const pendingApproval = pendingApprovals[0];
            let stub;
            if (coin.supportsTss()) {
                stub = sandbox.stub(sdk_core_1.PendingApproval.prototype, 'recreateAndSignTSSTransaction').resolves({
                    txHex: 'gabagool',
                });
            }
            else {
                stub = sandbox.stub(sdk_core_1.PendingApproval.prototype, 'recreateAndSignTransaction').resolves({
                    state: 'approved',
                    halfSigned: { txHex: 'gabagool' },
                });
            }
            const paScope = nock(bgUrl)
                .put(`/api/v2/${coinName}/pendingapprovals/${pendingApprovalDataTemp.id}`, {
                state: 'approved',
                halfSigned: type === sdk_core_1.Type.TRANSACTION_REQUEST ? { txHex: 'gabagool' } : undefined,
            })
                .reply(200, {
                ...pendingApprovalDataTemp,
                state: 'approved',
            });
            await pendingApproval.approve({ xprv: 'nonsense', walletPassphrase: 'gabagoolio' });
            // Should not call build and should call pa
            paScope.isDone().should.be.true();
            stub.calledOnce.should.equal(recreateTransaction);
        });
    }
    testRecreateTransaction('tbtc', false, sdk_core_1.Type.TRANSACTION_REQUEST);
    testRecreateTransaction('tsol', true, sdk_core_1.Type.TRANSACTION_REQUEST);
    testRecreateTransaction('tsol', true, sdk_core_1.Type.TRANSACTION_REQUEST_FULL);
    testRecreateTransaction('teth', true, sdk_core_1.Type.TRANSACTION_REQUEST_FULL);
    describe('recreateAndSignTSSTransaction', function () {
        let coin;
        before(() => {
            coin = bitgo.coin('tsol');
        });
        it('should call approve and do the TSS flow and fail if the txRequestId is missing', async () => {
            const pendingApproval = wallet.pendingApprovals()[0];
            const reqId = new sdk_core_1.RequestTracer();
            const params = { walletPassphrase: 'test' };
            await pendingApproval
                .recreateAndSignTSSTransaction(params, reqId)
                .should.be.rejectedWith('txRequestId not found');
        });
        it('should call approve and do the TSS flow and fail if the walletPassphrase is missing', async () => {
            const pendingApproval = wallet.pendingApprovals()[0];
            pendingApprovalData['txRequestId'] = 'requestTxIdTest';
            const reqId = new sdk_core_1.RequestTracer();
            const params = {};
            await pendingApproval
                .recreateAndSignTSSTransaction(params, reqId)
                .should.be.rejectedWith('walletPassphrase not found');
        });
        it('should call approve and do the TSS flow and fail if the wallet is missing', async () => {
            const pendingApproval = new sdk_core_1.PendingApproval(bitgo, coin, pendingApprovalData);
            const reqId = new sdk_core_1.RequestTracer();
            const params = { walletPassphrase: 'test' };
            await pendingApproval.recreateAndSignTSSTransaction(params, reqId).should.be.rejectedWith('Wallet not found');
        });
        it('should get txHex for transactionRequestLite', async () => {
            pendingApprovalData['txRequestId'] = 'requestTxIdTest';
            const pendingApproval = new sdk_core_1.PendingApproval(bitgo, coin, pendingApprovalData, wallet);
            const reqId = new sdk_core_1.RequestTracer();
            const txRequestId = 'test';
            const walletPassphrase = 'test';
            const decryptedPrvResponse = 'decryptedPrv';
            const params = { txRequestId, walletPassphrase };
            const txRequest = {
                apiVersion: 'lite',
                txRequestId: txRequestId,
                unsignedTxs: [{ signableHex: 'randomhex', serializedTxHex: 'randomhex2', derivationPath: 'm/0' }],
                signatureShares: [
                    {
                        from: sdk_core_1.SignatureShareType.BITGO,
                        to: sdk_core_1.SignatureShareType.USER,
                        share: '9d7159a76700635TEST',
                    },
                ],
                transactions: [],
                userId: 'userId',
                date: new Date().toISOString(),
                intent: {
                    intentType: 'payment',
                },
                latest: true,
                walletId: 'walletId',
                version: 1,
                policiesChecked: false,
                walletType: 'hot',
                state: 'pendingUserSignature',
            };
            const decryptedPrv = sandbox.stub(sdk_core_1.Wallet.prototype, 'getPrv');
            decryptedPrv.calledOnceWithExactly({ walletPassphrase });
            decryptedPrv.resolves(decryptedPrvResponse);
            const recreateTxRequest = sandbox.stub(sdk_core_1.TssUtils.prototype, 'recreateTxRequest');
            recreateTxRequest.calledOnceWithExactly(txRequest.txRequestId, decryptedPrvResponse, reqId);
            recreateTxRequest.resolves(txRequest);
            const recreatedTx = await pendingApproval.recreateAndSignTSSTransaction(params, reqId);
            recreatedTx.should.be.deepEqual({ txHex: txRequest.unsignedTxs[0].serializedTxHex });
            sandbox.verify();
        });
        it('should get txHex for transactionRequestFull ', async () => {
            pendingApprovalData['txRequestId'] = 'requestTxIdTest';
            const pendingApproval = new sdk_core_1.PendingApproval(bitgo, coin, pendingApprovalData, wallet);
            const reqId = new sdk_core_1.RequestTracer();
            const txRequestId = 'test';
            const walletPassphrase = 'test';
            const decryptedPrvResponse = 'decryptedPrv';
            const params = { txRequestId, walletPassphrase };
            const txRequest = {
                txRequestId: txRequestId,
                apiVersion: 'full',
                unsignedTxs: [],
                transactions: [
                    {
                        unsignedTx: { signableHex: 'randomhex', serializedTxHex: 'randomhex2', derivationPath: 'm/0' },
                        signatureShares: [
                            {
                                from: sdk_core_1.SignatureShareType.BITGO,
                                to: sdk_core_1.SignatureShareType.USER,
                                share: '9d7159a76700635TEST',
                            },
                        ],
                        state: 'initialized',
                    },
                ],
                userId: 'userId',
                date: new Date().toISOString(),
                intent: {
                    intentType: 'payment',
                },
                latest: true,
                walletId: 'walletId',
                version: 1,
                policiesChecked: false,
                walletType: 'hot',
                state: 'pendingDelivery',
            };
            const decryptedPrv = sandbox.stub(sdk_core_1.Wallet.prototype, 'getPrv');
            decryptedPrv.calledOnceWithExactly({ walletPassphrase });
            decryptedPrv.resolves(decryptedPrvResponse);
            const recreateTxRequest = sandbox.stub(sdk_core_1.TssUtils.prototype, 'recreateTxRequest');
            recreateTxRequest.calledOnceWithExactly(txRequest.txRequestId, decryptedPrvResponse, reqId);
            recreateTxRequest.resolves(txRequest);
            const recreatedTx = await pendingApproval.recreateAndSignTSSTransaction(params, reqId);
            recreatedTx.should.be.deepEqual({ txHex: txRequest.transactions[0].unsignedTx.serializedTxHex });
            sandbox.verify();
        });
    });
});
