// @bitgo/sdk-coin-ada/dist/src/lib/messages/cip8/cip8Message.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cip8Message = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const CardanoSL = __importStar(require("@emurgo/cardano-serialization-lib-nodejs"));
const utils_1 = require("./utils");
/**
 * Implementation of Message for CIP8 standard
 */
class Cip8Message extends sdk_core_1.BaseMessage {
    constructor(options) {
        super({
            ...options,
            type: sdk_core_1.MessageStandardType.CIP8,
        });
    }
    /**
     * Returns the hash of the CIP-8 prefixed message
     */
    async getSignablePayload() {
        const { addressCborBytes } = this.validateAndGetCommonSetup();
        const { sigStructureCborBytes } = (0, utils_1.createCSLSigStructure)(addressCborBytes, this.payload);
        this.signablePayload = Buffer.from(sigStructureCborBytes);
        return this.signablePayload;
    }
    /*
     * Returns broadcastable signatures in COSE format according to CIP8 standard
     *
     * This method transforms the internal signatures into a format suitable for broadcasting
     * by constructing COSE (CBOR Object Signing and Encryption) objects that comply with
     * the CIP8 message signing specification.
     *
     * @returns Array of signatures with COSE-formatted signature data and public keys
     * @throws Error if required setup validation fails
     */
    getBroadcastableSignatures() {
        if (!this.signatures.length) {
            return [];
        }
        const signature = this.signatures[0].signature;
        const publicKeyHex = this.signatures[0].publicKey.pub;
        const { addressCborBytes } = this.validateAndGetCommonSetup();
        const { protectedHeaderCborBytes, payloadBytes } = (0, utils_1.createCSLSigStructure)(addressCborBytes, this.payload);
        const coseObjectsOutput = (0, utils_1.constructCSLCoseObjects)(protectedHeaderCborBytes, payloadBytes, signature, CardanoSL.PublicKey.from_bytes(Buffer.from(publicKeyHex, 'hex')));
        const coseObjectsBuffer = (0, utils_1.coseObjectsOutputToBuffer)(coseObjectsOutput);
        return [
            {
                signature: coseObjectsBuffer,
                publicKey: {
                    pub: publicKeyHex,
                },
            },
        ];
    }
    /**
     * Verifies the encoded payload against the provided metadata
     * @param messageEncodedHex The hex-encoded message to verify
     * @param metadata Metadata containing signer addresses
     * @returns True if the encoded payload matches the expected format, false otherwise
     */
    async verifyEncodedPayload(messageEncodedHex, metadata) {
        if (!metadata) {
            throw new Error('Metadata is required for verifying the encoded payload');
        }
        const signers = metadata.signers;
        if (signers.length === 0) {
            throw new Error('At least one signer address is required in metadata for verification');
        }
        this.addSigner(signers[0]);
        const signablePayload = await this.getSignablePayload();
        let signablePayloadHex;
        if (Buffer.isBuffer(signablePayload)) {
            signablePayloadHex = signablePayload.toString('hex');
        }
        else {
            signablePayloadHex = signablePayload;
        }
        return signablePayloadHex === messageEncodedHex;
    }
    /**
     * Validates required fields and returns common setup objects
     * @private
     */
    validateAndGetCommonSetup() {
        if (!this.payload) {
            throw new Error('Payload is required to build a CIP8 message');
        }
        if (!this.signers || this.signers.length === 0) {
            throw new Error('A signer address is required to build a CIP8 message');
        }
        let cslAddress;
        try {
            cslAddress = CardanoSL.Address.from_bech32(this.signers[0]);
        }
        catch (error) {
            // Convert string errors to proper Error objects
            if (typeof error === 'string') {
                throw new Error(`Invalid signer address: ${error}`);
            }
            throw error;
        }
        const addressCborBytes = cslAddress.to_bytes();
        return { addressCborBytes };
    }
}
exports.Cip8Message = Cip8Message;
