// @bitgo/sdk-coin-ada/dist/src/lib/messages/cip8/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bytesToHex = bytesToHex;
exports.createCSLSigStructure = createCSLSigStructure;
exports.constructCSLCoseObjects = constructCSLCoseObjects;
exports.coseObjectsOutputToBuffer = coseObjectsOutputToBuffer;
exports.bufferToCoseObjectsOutput = bufferToCoseObjectsOutput;
const buffer_1 = require("buffer");
const cbor_1 = require("cbor");
// Helper function to convert a Uint8Array or Buffer to a hex string
function bytesToHex(bytes) {
    return buffer_1.Buffer.from(bytes).toString('hex');
}
/**
 * Creates the CSL signature structure for off-chain message signing.
 *
 * @param addressCborBytes - The CBOR bytes of the CSL address.
 * @param message - The message to be signed.
 * @returns An object containing the signature structure CBOR bytes, protected header CBOR bytes, and payload bytes.
 */
function createCSLSigStructure(addressCborBytes, message) {
    // Payload
    const payloadBytes = buffer_1.Buffer.from(message, 'utf-8');
    // Protected Header
    const protectedHeaderMap = new Map();
    protectedHeaderMap.set(1, -8); // Algorithm ID: EdDSA
    protectedHeaderMap.set('address', buffer_1.Buffer.from(addressCborBytes));
    const protectedHeaderCborBytes = cbor_1.Encoder.encode(protectedHeaderMap);
    // Sig_structure
    const sigStructureArray = [
        'Signature1',
        buffer_1.Buffer.from(protectedHeaderCborBytes),
        buffer_1.Buffer.from([]), // Empty external_aad
        buffer_1.Buffer.from(payloadBytes),
    ];
    const sigStructureCborBytes = cbor_1.Encoder.encode(sigStructureArray);
    return { sigStructureCborBytes, protectedHeaderCborBytes, payloadBytes };
}
// COSE objects construction function
function constructCSLCoseObjects(protectedHeaderCborBytes, payloadBytes, cslSignatureBytes, paymentPubKey) {
    // COSE_Sign1 Construction
    const unprotectedHeadersMap = new Map();
    unprotectedHeadersMap.set('hashed', false);
    const coseSign1Array = [
        buffer_1.Buffer.from(protectedHeaderCborBytes),
        unprotectedHeadersMap,
        buffer_1.Buffer.from(payloadBytes),
        buffer_1.Buffer.from(cslSignatureBytes),
    ];
    const finalCoseSign1CborBytes = cbor_1.Encoder.encode(coseSign1Array);
    /* // directly encoding the coseSign1Array without prepending the 0xD2 tag.
     * const coseSign1PayloadBytes = Encoder.encode(coseSign1Array);
     * const coseSign1Tag = Buffer.from([0xD2]); // Tag 18 for COSE_Sign1
     * const finalCoseSign1CborBytes = Buffer.concat([coseSign1Tag, coseSign1PayloadBytes]);
     */
    const manualCoseSign1Hex = bytesToHex(finalCoseSign1CborBytes);
    // COSE_Key Construction
    const coseKeyMap = new Map();
    coseKeyMap.set(1, 1); // kty: OKP (Octet Key Pair)
    coseKeyMap.set(3, -8); // alg: EdDSA
    coseKeyMap.set(-1, 6); // crv: Ed25519
    coseKeyMap.set(-2, buffer_1.Buffer.from(paymentPubKey.as_bytes())); // x: public_key_bytes (Ed25519 public key)
    const finalCoseKeyCborBytes = cbor_1.Encoder.encode(coseKeyMap);
    const manualCoseKeyHex = bytesToHex(finalCoseKeyCborBytes);
    return { manualCoseSign1Hex, manualCoseKeyHex };
}
function coseObjectsOutputToBuffer(output) {
    return buffer_1.Buffer.from(cbor_1.Encoder.encode(output));
}
async function bufferToCoseObjectsOutput(buffer) {
    return await cbor_1.Decoder.decodeFirst(buffer);
}
