// @bitgo/sdk-core/dist/src/bitgo/proofs/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountSnapshot = exports.SnapshotBalance = exports.UserVerificationElements = exports.TopProof = exports.LowerProof = exports.BalancesCodec = void 0;
const t = __importStar(require("io-ts"));
// type definitions for user verification elements (returned by liability proofs route)
exports.BalancesCodec = t.type({
    Asset: t.string,
    Amount: t.string,
});
exports.LowerProof = t.type({
    Proof: t.string,
    VerificationKey: t.string,
    MerkleRoot: t.string,
    MerkleRootWithAssetSumHash: t.string,
    MerklePath: t.array(t.string),
    MerklePosition: t.number,
});
exports.TopProof = t.type({
    Proof: t.string,
    VerificationKey: t.string,
    MerkleRoot: t.string,
    MerkleRootWithAssetSumHash: t.string,
    AssetSum: t.array(exports.BalancesCodec),
});
exports.UserVerificationElements = t.type({
    AccountInfo: t.type({
        WalletId: t.string,
        Balance: t.array(exports.BalancesCodec),
    }),
    ProofInfo: t.type({
        UserMerklePath: t.array(t.string),
        UserMerklePosition: t.number,
        BottomProof: exports.LowerProof,
        MiddleProof: exports.LowerProof,
        TopProof: exports.TopProof,
    }),
});
// type definitions for account snapshots
exports.SnapshotBalance = t.type({
    asset: t.string,
    amount: t.string,
});
exports.AccountSnapshot = t.type({
    snapshotDate: t.string,
    balances: t.array(exports.SnapshotBalance),
});
