// @bitgo/abstract-utxo/dist/cjs/src/offlineVault/descriptor/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescriptorTransaction = void 0;
exports.getDescriptorsFromDescriptorTransaction = getDescriptorsFromDescriptorTransaction;
exports.getHalfSignedPsbt = getHalfSignedPsbt;
exports.getTransactionExplanationFromPsbt = getTransactionExplanationFromPsbt;
const t = __importStar(require("io-ts"));
const descriptor_1 = require("../../descriptor");
const OfflineVaultSignable_1 = require("../OfflineVaultSignable");
const validatePolicy_1 = require("../../descriptor/validatePolicy");
const descriptor_2 = require("../../transaction/descriptor");
const wasmUtil_1 = require("../../wasmUtil");
exports.DescriptorTransaction = t.intersection([OfflineVaultSignable_1.OfflineVaultSignable, t.type({ descriptors: t.array(descriptor_1.NamedDescriptor) })], 'DescriptorTransaction');
function getDescriptorsFromDescriptorTransaction(tx) {
    const { descriptors, xpubsWithDerivationPath } = tx;
    const pubkeys = (0, OfflineVaultSignable_1.toKeyTriple)(xpubsWithDerivationPath);
    const policy = (0, validatePolicy_1.getValidatorSome)([
        // allow all 2-of-3-ish descriptors where the keys match the wallet keys
        (0, validatePolicy_1.getValidatorOneOfTemplates)(['Wsh2Of3', 'Wsh2Of3CltvDrop', 'ShWsh2Of3CltvDrop']),
        // allow all descriptors signed by the user key
        (0, validatePolicy_1.getValidatorSignedByUserKey)(),
    ]);
    return (0, validatePolicy_1.toDescriptorMapValidate)(descriptors, pubkeys, policy);
}
function getHalfSignedPsbt(tx, prv, coinName) {
    const psbt = (0, wasmUtil_1.toWasmPsbt)(Buffer.from(tx.coinSpecific.txHex, 'hex'));
    const descriptorMap = getDescriptorsFromDescriptorTransaction(tx);
    (0, descriptor_2.signPsbt)(psbt, descriptorMap, prv, { onUnknownInput: 'throw' });
    return psbt;
}
function getTransactionExplanationFromPsbt(tx, coinName) {
    const psbt = (0, wasmUtil_1.toWasmPsbt)(Buffer.from(tx.coinSpecific.txHex, 'hex'));
    const descriptorMap = getDescriptorsFromDescriptorTransaction(tx);
    const { outputs, changeOutputs, fee } = (0, descriptor_2.explainPsbt)(psbt, descriptorMap, coinName);
    return {
        outputs,
        changeOutputs,
        fee: {
            /* network fee */
            fee,
            /* TODO */
            payGoFeeString: undefined,
            payGoFeeAddress: undefined,
        },
    };
}
