// @bitgo/sdk-core/dist/src/coins/fiatinr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FiatINR = void 0;
/**
 * @prettier
 */
const __1 = require("../");
class FiatINR extends __1.BaseCoin {
    static createInstance(bitgo) {
        return new FiatINR(bitgo);
    }
    /**
     * Returns the factor between the base unit and its smallest subdivison
     * @return {number}
     */
    getBaseFactor() {
        return 1e2;
    }
    getChain() {
        return 'fiatinr';
    }
    getFamily() {
        return 'fiat';
    }
    getFullName() {
        return 'Indian Rupee';
    }
    /**
     * Return whether the given m of n wallet signers/ key amounts are valid for the coin
     */
    isValidMofNSetup({ m, n }) {
        return m === 0 && n === 0;
    }
    isValidAddress(address) {
        throw new __1.MethodNotImplementedError();
    }
    generateKeyPair(seed) {
        throw new __1.MethodNotImplementedError();
    }
    isValidPub(pub) {
        throw new __1.MethodNotImplementedError();
    }
    async parseTransaction(params) {
        return {};
    }
    async isWalletAddress(params) {
        throw new __1.MethodNotImplementedError();
    }
    async verifyTransaction(params) {
        return true;
    }
    async signTransaction(params = {}) {
        throw new __1.MethodNotImplementedError();
    }
    /** @inheritDoc */
    auditDecryptedKey(params) {
        throw new __1.MethodNotImplementedError();
    }
}
exports.FiatINR = FiatINR;
