// @bitgo/sdk-core/dist/src/coins/ofcToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfcToken = void 0;
const lodash_1 = require("lodash");
const lightning_1 = require("../lightning");
const ofc_1 = require("./ofc");
const publicIdRegex = /^[a-f\d]{32}$/i;
class OfcToken extends ofc_1.Ofc {
    constructor(bitgo, tokenConfig) {
        super(bitgo);
        this.tokenConfig = tokenConfig;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get backingCoin() {
        return this.tokenConfig.backingCoin;
    }
    get isFiat() {
        return this.tokenConfig.isFiat;
    }
    getChain() {
        return this.type;
    }
    getFullName() {
        return this.name;
    }
    getBaseFactor() {
        return String(Math.pow(10, this.decimalPlaces));
    }
    get type() {
        return this.tokenConfig.type;
    }
    checkRecipient(recipient) {
        if ((0, lightning_1.isBolt11Invoice)(recipient.address)) {
            // should throw error if this isnt bitcoin (mainnet or testnet)
            if (this.backingCoin !== 'btc' && this.backingCoin !== 'tbtc') {
                throw new Error(`invalid argument - lightning invoice is only supported for bitcoin`);
            }
            // amount for bolt11 invoices is either 'invoice' or a non-zero bigint
            if (recipient.amount === lightning_1.LIGHTNING_INVOICE) {
                return;
            }
            // try to parse the amount as a bigint
            let amount;
            try {
                amount = BigInt(recipient.amount);
            }
            catch (e) {
                throw new Error(`invalid argument ${recipient.amount} for amount - lightning invoice amount must be >= 0 or ${lightning_1.LIGHTNING_INVOICE}`);
            }
            if (amount > 0n) {
                return;
            }
            throw new Error(`invalid argument for amount - lightning invoice amount must be a non-zero bigint or ${lightning_1.LIGHTNING_INVOICE}`);
        }
        super.checkRecipient(recipient);
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return false;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new OfcToken(bitgo, config);
    }
    /**
     * Assemble keychain and half-sign prebuilt transaction
     * @param params
     * @returns {Promise<SignedTransaction>}
     */
    async signTransaction(params) {
        const txPrebuild = params.txPrebuild;
        const payload = txPrebuild.payload;
        const signatureBuffer = (await this.signMessage(params, payload));
        const signature = signatureBuffer.toString('hex');
        return { halfSigned: { payload, signature } };
    }
    /**
     * Check if an address is valid for this ofc token.
     *
     * These addresses are either bg-<publicid>, where public id is the internal address to send to,
     * or are an address which is valid on the backing coin of this ofc token.
     * @param address address to check for validity
     */
    isValidAddress(address) {
        if (!(0, lodash_1.isString)(address)) {
            return false;
        }
        if (address.startsWith('bg-')) {
            const parts = address.split('-');
            const accountId = parts[1];
            return parts.length === 2 && publicIdRegex.test(accountId);
        }
        else {
            const backingCoin = this.bitgo.coin(this.backingCoin);
            return backingCoin.isValidAddress(address);
        }
    }
}
exports.OfcToken = OfcToken;
