// @bitgo/sdk-core/dist/src/bitgo/walletUtil/midnightMessageProvider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MidnightMessageProvider = exports.MIDNIGHT_TNC_HASH = void 0;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const environments_1 = require("../environments");
const NUM_MESSAGES_PER_QUERY = 1000;
exports.MIDNIGHT_TNC_HASH = '31a6bab50a84b8439adcfb786bb2020f6807e6e8fda629b424110fc7bb1c6b8b';
/**
 * The Midnight drop service can return up to 1000 messages per request.
 * We make this wrapper function that handles the pagination and batching of messages,
 * keeping a local cache of the unprocessed messages.
 */
class MidnightMessageProvider {
    constructor(wallet, destinationAddress, batchSize = NUM_MESSAGES_PER_QUERY) {
        this.wallet = wallet;
        this.destinationAddress = destinationAddress;
        this.batchSize = batchSize;
        this.ranOnce = false;
        this.unprocessedMessagesCache = [];
        this.network = utxolib.networks[wallet.coin()];
        this.midnightClaimUrl = `${environments_1.Environments[wallet.bitgo.env].uri}/api/airdrop-claim/v1/midnight/claims/${wallet.coin()}/${wallet.id()}`;
    }
    async getMessagesAndAddressesToSign() {
        if (this.unprocessedMessagesCache.length > 0) {
            return this.unprocessedMessagesCache.splice(0, this.batchSize);
        }
        else if (this.unprocessedMessagesCache.length === 0 && this.ranOnce && this.prevId === undefined) {
            return [];
        }
        this.ranOnce = true;
        const query = {
            statuses: ['UNINITIATED', 'NEEDS_RESUBMITTING'],
            limit: NUM_MESSAGES_PER_QUERY,
        };
        if (this.prevId !== undefined) {
            query.prevId = this.prevId;
        }
        const response = await this.wallet.bitgo.get(this.midnightClaimUrl).query(query).result();
        if (response.status !== 'success') {
            throw new Error(`Unexpected status code ${response.status} from ${this.midnightClaimUrl}`);
        }
        if (response?.pagination?.hasNext) {
            this.prevId = response?.pagination?.nextPrevId;
        }
        else {
            this.prevId = undefined;
        }
        this.unprocessedMessagesCache = response.claims.map((claim) => {
            if (!claim.originAddress) {
                throw new Error(`Claim ${JSON.stringify(claim)} is missing originAddress`);
            }
            return {
                // Midnight claim message format
                message: `STAR ${claim.allocationAmount} to ${this.destinationAddress} ${exports.MIDNIGHT_TNC_HASH}`,
                address: claim.originAddress,
            };
        });
        return this.unprocessedMessagesCache.splice(0, this.batchSize);
    }
}
exports.MidnightMessageProvider = MidnightMessageProvider;
