// @bitgo/sdk-core/dist/src/bitgo/walletUtil/signAccountBasedMidnightClaimMessages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bulkSignAccountBasedMidnightClaimMessages = bulkSignAccountBasedMidnightClaimMessages;
const wallet_1 = require("../wallet");
const midnightMessageProvider_1 = require("./midnightMessageProvider");
async function bulkSignAccountBasedMidnightClaimMessages(wallet, messageStandardType, destinationAddress, walletPassphrase) {
    const provider = new midnightMessageProvider_1.MidnightMessageProvider(wallet, destinationAddress);
    return bulkSignAccountBasedMessagesWithProvider(provider, wallet, messageStandardType, walletPassphrase);
}
async function bulkSignAccountBasedMessagesWithProvider(provider, wallet, messageStandardType, walletPassphrase) {
    const failedAddresses = [];
    const txRequests = [];
    // Gather all messages to process (flatten all batches)
    let allMessages = [];
    let messages = await provider.getMessagesAndAddressesToSign();
    while (messages.length > 0) {
        allMessages = allMessages.concat(messages);
        messages = await provider.getMessagesAndAddressesToSign();
    }
    // Extract wallet constructor params
    const { bitgo, baseCoin, _wallet: walletData } = wallet;
    // Process all messages sequentially, each with a new Wallet instance
    const results = [];
    for (const messageInfo of allMessages) {
        const newWallet = new wallet_1.Wallet(bitgo, baseCoin, walletData);
        const result = await signOrBuildMessage(newWallet, messageInfo, messageStandardType, walletPassphrase);
        results.push(result);
    }
    processResults(results, txRequests, failedAddresses);
    return { failedAddresses, txRequests };
}
async function signOrBuildMessage(wallet, messageInfo, messageStandardType, walletPassphrase) {
    let txRequestId;
    if (walletPassphrase !== undefined) {
        // Sign the messages with the wallet
        const signedMessage = await wallet.signMessage({
            message: {
                messageRaw: messageInfo.message,
                messageStandardType,
                signerAddress: messageInfo.address,
            },
            walletPassphrase,
        });
        txRequestId = signedMessage.txRequestId;
    }
    else {
        // Build the sign message request
        const txRequest = await wallet.buildSignMessageRequest({
            message: {
                messageRaw: messageInfo.message,
                messageStandardType,
                signerAddress: messageInfo.address,
            },
        });
        txRequestId = txRequest.txRequestId;
    }
    return { success: true, address: messageInfo.address, txRequestId };
}
function processResults(results, txRequests, failedAddresses) {
    for (const result of results) {
        if (result.success) {
            txRequests.push({
                address: result.address,
                txRequestId: result.txRequestId,
            });
        }
        else {
            failedAddresses.push(result.address);
        }
    }
}
