// @bitgo/sdk-coin-canton/dist/src/lib/transaction/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
const constant_1 = require("../constant");
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
    }
    get prepareCommand() {
        return this._prepareCommand;
    }
    set prepareCommand(transaction) {
        this._prepareCommand = transaction;
    }
    set transactionType(transactionType) {
        this._type = transactionType;
    }
    set acknowledgeData(data) {
        this._acknowledgeData = data;
    }
    get id() {
        if (!this._id) {
            throw new sdk_core_1.InvalidTransactionError('transaction is is not set');
        }
        return this._id;
    }
    set id(id) {
        this._id = id;
    }
    canSign(key) {
        return false;
    }
    set signatures(signature) {
        this._signatures.push(signature);
    }
    set signerFingerprint(fingerprint) {
        this._signerFingerprint = fingerprint;
    }
    toBroadcastFormat() {
        if (this._type === undefined) {
            throw new sdk_core_1.InvalidTransactionError('Transaction type is not set');
        }
        if (this._type === sdk_core_1.TransactionType.TransferAcknowledge) {
            if (!this._acknowledgeData) {
                throw new sdk_core_1.InvalidTransactionError('AcknowledgeData is not set');
            }
            const minData = {
                txType: sdk_core_1.TransactionType[this._type],
                submissionId: this.id,
                acknowledgeData: this._acknowledgeData,
            };
            return Buffer.from(JSON.stringify(minData)).toString('base64');
        }
        if (!this._prepareCommand) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        const partySignatures = [];
        const data = {
            prepareCommandResponse: this._prepareCommand,
            txType: this._type !== undefined ? sdk_core_1.TransactionType[this._type] : '',
            preparedTransaction: '',
            partySignatures: {
                signatures: partySignatures,
            },
            deduplicationPeriod: {
                Empty: {},
            },
            submissionId: this.id,
            hashingSchemeVersion: constant_1.HASHING_SCHEME_VERSION,
            minLedgerTime: {
                time: {
                    Empty: {},
                },
            },
        };
        const signatures = [];
        if (this._signatures.length > 0 && this._signerFingerprint) {
            const signerPartyId = `${this._signerFingerprint.slice(0, 5)}::${this._signerFingerprint}`;
            this.signature.map((signature) => {
                const signatureObj = {
                    format: constant_1.SIGNATURE_FORMAT,
                    signature: signature,
                    signedBy: this._signerFingerprint,
                    signingAlgorithmSpec: constant_1.SIGNATURE_ALGORITHM_SPEC,
                };
                signatures.push(signatureObj);
            });
            const partySignature = {
                party: signerPartyId,
                signatures: signatures,
            };
            data.partySignatures?.signatures.push(partySignature);
            data.preparedTransaction = this._prepareCommand.preparedTransaction
                ? this._prepareCommand.preparedTransaction
                : '';
        }
        return Buffer.from(JSON.stringify(data)).toString('base64');
    }
    toJson() {
        const result = {
            id: this.id,
            type: this._type,
            sender: '',
            receiver: '',
            amount: '',
        };
        if (this._type === sdk_core_1.TransactionType.TransferAcknowledge) {
            if (!this._acknowledgeData) {
                throw new sdk_core_1.InvalidTransactionError('AcknowledgeData is not set');
            }
            result.acknowledgeData = this._acknowledgeData;
            return result;
        }
        if (!this._prepareCommand || !this._prepareCommand.preparedTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        // TODO: extract other required data (utxo used, request time, execute before etc)
        let parsedInfo;
        try {
            parsedInfo = utils_1.default.parseRawCantonTransactionData(this._prepareCommand.preparedTransaction, this.type);
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError(`Failed to parse transaction hash: ${e instanceof Error ? e.message : e}`);
        }
        result.sender = parsedInfo.sender;
        result.receiver = parsedInfo.receiver;
        result.amount = parsedInfo.amount;
        if (parsedInfo.memoId) {
            result.memoId = parsedInfo.memoId;
        }
        if (parsedInfo.token) {
            result.token = parsedInfo.token;
        }
        return result;
    }
    get signablePayload() {
        if (this._type === sdk_core_1.TransactionType.TransferAcknowledge) {
            return Buffer.from(constant_1.DUMMY_HASH, 'base64');
        }
        if (!this._prepareCommand) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        return Buffer.from(this._prepareCommand.preparedTransactionHash, 'base64');
    }
    fromRawTransaction(rawTx) {
        try {
            const decoded = JSON.parse(Buffer.from(rawTx, 'base64').toString('utf8'));
            this.id = decoded.submissionId;
            this.transactionType = sdk_core_1.TransactionType[decoded.txType];
            if (this.type !== sdk_core_1.TransactionType.TransferAcknowledge) {
                if (decoded.prepareCommandResponse) {
                    this.prepareCommand = decoded.prepareCommandResponse;
                    this.loadInputsAndOutputs();
                }
                if (decoded.partySignatures && decoded.partySignatures.signatures.length > 0) {
                    this.signerFingerprint = decoded.partySignatures.signatures[0].party.split('::')[1];
                    this.signatures = decoded.partySignatures.signatures[0].signatures[0].signature;
                }
            }
            else {
                if (decoded.acknowledgeData) {
                    this.acknowledgeData = decoded.acknowledgeData;
                }
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError('Unable to parse raw transaction data');
        }
    }
    /**
     * Loads the input & output fields for the transaction
     *
     */
    loadInputsAndOutputs() {
        const outputs = [];
        const inputs = [];
        const txData = this.toJson();
        const input = {
            address: txData.sender,
            value: txData.amount,
            coin: txData.token ? txData.token : this._coinConfig.name,
        };
        const output = {
            address: txData.receiver,
            value: txData.amount,
            coin: txData.token ? txData.token : this._coinConfig.name,
        };
        inputs.push(input);
        outputs.push(output);
        this._inputs = inputs;
        this._outputs = outputs;
    }
    explainTransaction() {
        const displayOrder = [
            'id',
            'outputs',
            'outputAmount',
            'inputs',
            'inputAmount',
            'changeOutputs',
            'changeAmount',
            'fee',
            'type',
        ];
        const inputs = [];
        const outputs = [];
        let inputAmount = '0';
        let outputAmount = '0';
        switch (this.type) {
            case sdk_core_1.TransactionType.TransferAccept:
            case sdk_core_1.TransactionType.TransferReject: {
                const txData = this.toJson();
                const input = {
                    address: txData.sender,
                    amount: txData.amount,
                };
                if (txData.token) {
                    input.tokenName = txData.token;
                }
                inputs.push(input);
                inputAmount = txData.amount;
                break;
            }
            case sdk_core_1.TransactionType.Send: {
                const txData = this.toJson();
                const output = {
                    address: txData.receiver,
                    amount: txData.amount,
                };
                if (txData.memoId) {
                    output.memo = txData.memoId;
                }
                if (txData.token) {
                    output.tokenName = txData.token;
                }
                outputs.push(output);
                outputAmount = txData.amount;
                break;
            }
            case sdk_core_1.TransactionType.TransferOfferWithdrawn: {
                const txData = this.toJson();
                const input = {
                    address: txData.receiver,
                    amount: txData.amount,
                };
                if (txData.token) {
                    input.tokenName = txData.token;
                }
                inputs.push(input);
                inputAmount = txData.amount;
                break;
            }
        }
        return {
            id: this.id,
            displayOrder,
            outputs: outputs,
            outputAmount: outputAmount,
            inputs: inputs,
            inputAmount: inputAmount,
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: '0' },
            type: this.type,
        };
    }
}
exports.Transaction = Transaction;
