// @bitgo/abstract-utxo/dist/cjs/test/integration/impl/bch/bch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const should = __importStar(require("should"));
require("should-http");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const nock = __importStar(require("nock"));
const bch_1 = require("../../../../src/impl/bch");
nock.restore();
describe('BCH:', function () {
    let bitgo;
    let wallet;
    before(async function () {
        this.timeout(10000);
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
        bitgo.initializeTestVars();
        bitgo.safeRegister('tbch', bch_1.Tbch.createInstance);
        await bitgo.authenticateTestUser(bitgo.testUserOTP());
        wallet = await bitgo.coin('tbch').wallets().getWallet({ id: sdk_test_1.TestBitGo.V2.TEST_BCH_WALLET_ID });
    });
    describe('Send Transaction', function () {
        it('should send fund to cashaddr recipient', async function () {
            this.timeout(10000);
            await bitgo.unlock({ otp: '0000000' });
            const transaction = await wallet.send({
                address: sdk_test_1.TestBitGo.V2.TEST_BCH_WALLET_CASH_ADDRESS,
                amount: '5000',
                walletPassphrase: sdk_test_1.TestBitGo.V2.TEST_BCH_WALLET_PASSPHRASE,
                otp: bitgo.testUserOTP(),
            });
            should.exist(transaction);
            transaction.should.have.property('transfer');
            transaction.should.have.property('txid');
            transaction.should.have.property('tx');
            transaction.status.should.containEql('signed');
            transaction.transfer.type.should.containEql('send');
            transaction.transfer.wallet.should.containEql(wallet._wallet.id);
        });
    });
});
