// @bitgo/sdk-coin-ethlike/dist/src/ethlikeCoin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthLikeCoin = void 0;
const common_1 = __importDefault(require("@ethereumjs/common"));
const abstract_eth_1 = require("@bitgo/abstract-eth");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
class EthLikeCoin extends abstract_eth_1.AbstractEthLikeNewCoins {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
    }
    static createInstance(bitgo, staticsCoin) {
        return new EthLikeCoin(bitgo, staticsCoin);
    }
    getTransactionBuilder(common) {
        let ethereumCommon;
        // If common is an EthereumCommon instance, use it directly
        if (common instanceof common_1.default) {
            ethereumCommon = common;
        }
        // If common is provided as a plain object, convert it to EthereumCommon instance
        else if (common && typeof common === 'object') {
            try {
                const chainId = common.chain || common.chainId;
                const hardfork = common.hardfork || 'london';
                if (chainId) {
                    ethereumCommon = common_1.default.custom({
                        name: this.getFullName(),
                        chainId: chainId,
                        defaultHardfork: hardfork,
                    });
                }
            }
            catch (error) {
                ethereumCommon = undefined;
            }
        }
        return new lib_1.EthLikeTransactionBuilder(statics_1.coins.get(this.getBaseChain()), ethereumCommon);
    }
    async recoveryBlockchainExplorerQuery(query, apiKey) {
        const apiToken = apiKey ?? sdk_core_1.common.Environments[this.bitgo.getEnv()][this.getFamily().toLowerCase() + 'ApiToken'];
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()][this.getFamily().toLowerCase() + 'ExplorerUrl'];
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
}
exports.EthLikeCoin = EthLikeCoin;
