// @bitgo/public-types/dist/src/schema/lightning/payment.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightningPaymentRequestWithSig = exports.LightningPaymentRequest = exports.LightningOnchainRequest = exports.LightningOnchainRecipient = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
exports.LightningOnchainRecipient = t.type({
    amountSat: io_ts_types_1.BigIntFromString,
    address: t.string,
});
exports.LightningOnchainRequest = t.intersection([
    t.type({
        recipients: t.array(exports.LightningOnchainRecipient),
    }),
    t.partial({
        satsPerVbyte: io_ts_types_1.BigIntFromString,
        numBlocks: t.number,
    }),
]);
exports.LightningPaymentRequest = t.intersection([
    t.type({
        invoice: t.string,
    }),
    t.partial({
        amountMsat: io_ts_types_1.BigIntFromString,
        feeLimitMsat: io_ts_types_1.BigIntFromString,
        feeLimitRatio: t.number,
    }),
]);
exports.LightningPaymentRequestWithSig = t.type({
    signedRequest: exports.LightningPaymentRequest,
    signature: t.string,
});
