// @bitgo/sdk-coin-stx/dist/test/unit/transactionBuilder/contractBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const bn_js_1 = __importDefault(require("bn.js"));
const network_1 = require("@stacks/network");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactions_1 = require("@stacks/transactions");
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const statics_1 = require("@bitgo/statics");
const src_1 = require("../../../src");
const testData = __importStar(require("../resources"));
const { stringifyCv } = src_1.StxLib.Utils;
describe('Stacks: Contract Builder', function () {
    const coinName = 'stx';
    const coinNameTest = 'tstx';
    let bitgo;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, {
            env: 'mock',
        });
        bitgo.initializeTestVars();
        bitgo.safeRegister('stx', src_1.Stx.createInstance);
        bitgo.safeRegister('tstx', src_1.Tstx.createInstance);
    });
    describe('Stx Contract call Builder', () => {
        const factory = new src_1.StxLib.TransactionBuilderFactory(statics_1.coins.get(coinNameTest));
        const factoryProd = new src_1.StxLib.TransactionBuilderFactory(statics_1.coins.get(coinName));
        const initTxBuilder = () => {
            const txBuilder = factory.getContractBuilder();
            txBuilder.fee({ fee: '180' });
            txBuilder.nonce(0);
            txBuilder.contractAddress(testData.CONTRACT_ADDRESS);
            txBuilder.contractName(testData.CONTRACT_NAME);
            txBuilder.functionName(testData.CONTRACT_FUNCTION_NAME);
            return txBuilder;
        };
        describe('contract builder environment', function () {
            it('should select the right network', function () {
                should_1.default.equal(factory.getTransferBuilder().coinName(), 'tstx');
                should_1.default.equal(factoryProd.getTransferBuilder().coinName(), 'stx');
                // used type any to access protected properties
                const txBuilder = factory.getTransferBuilder();
                const txBuilderProd = factoryProd.getTransferBuilder();
                txBuilder._network.should.deepEqual(new network_1.StacksTestnet());
                txBuilderProd._network.should.deepEqual(new network_1.StacksMainnet());
            });
        });
        describe('should build ', () => {
            it('an unsigned contract call transaction', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([
                    { type: 'uint128', val: '400000000' },
                    { type: 'principal', val: testData.ACCOUNT_2.address },
                    { type: 'optional', val: { type: 'uint128', val: '200' } },
                    {
                        type: 'optional',
                        val: {
                            type: 'tuple',
                            val: [
                                { key: 'hashbytes', type: 'buffer', val: Buffer.from('some-hash') },
                                { key: 'version', type: 'buffer', val: new bn_js_1.default(1).toBuffer() },
                            ],
                        },
                    },
                ]);
                builder.fromPubKey(testData.TX_SENDER.pub);
                builder.numberSignatures(1);
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.UNSIGNED_CONTRACT_CALL);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.CONTRACT_ADDRESS);
                tx.outputs[0].value.should.equal('0');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.TX_SENDER.address);
                tx.inputs[0].value.should.equal('0');
            });
            it('an unsigned self stacking contract call transaction', async () => {
                const builder = initTxBuilder();
                /* Contract call in clarity POX-4
                (define-public (stack-stx (amount-ustx uint)
                                  (pox-addr (tuple (version (buff 1)) (hashbytes (buff 32))))
                                  (start-burn-ht uint)
                                  (lock-period uint)
                                  (signer-sig (optional (buff 65)))
                                  (signer-key (buff 33))
                                  (max-amount uint)
                                  (auth-id uint))
                 */
                builder.functionArgs([
                    { type: 'uint128', val: '400000000' },
                    {
                        type: 'tuple',
                        val: [
                            { key: 'hashbytes', type: 'buffer', val: Buffer.from('some-hash') },
                            { key: 'version', type: 'buffer', val: new bn_js_1.default(1).toBuffer() },
                        ],
                    },
                    { type: 'uint128', val: '52800' },
                    { type: 'uint128', val: '2' },
                    // Nakamoto upgrade new 4 parameters
                    // https://docs.stacks.co/nakamoto-upgrade/signing-and-stacking/stacking-flow#solo-stacker-flow
                    { type: 'optional', val: { type: 'buffer', val: Buffer.from('some-hash') } },
                    { type: 'buffer', val: Buffer.from('some-hash') },
                    { type: 'uint128', val: '340282366920938463463374607431768211455' },
                    { type: 'uint128', val: '123456' },
                ]);
                builder.fromPubKey(testData.TX_SENDER.pub);
                builder.numberSignatures(1);
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.UNSIGNED_SELF_STACK_CONTRACT_CALL);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.CONTRACT_ADDRESS);
                tx.outputs[0].value.should.equal('0');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.TX_SENDER.address);
                tx.inputs[0].value.should.equal('0');
            });
            it('a signed contract call with args', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([
                    { type: 'uint128', val: '400000000' },
                    { type: 'principal', val: testData.ACCOUNT_2.address },
                    { type: 'optional' },
                    {
                        type: 'optional',
                        val: {
                            type: 'tuple',
                            val: [
                                { key: 'hashbytes', type: 'buffer', val: Buffer.from('some-hash') },
                                { key: 'version', type: 'buffer', val: new bn_js_1.default(1).toBuffer() },
                            ],
                        },
                    },
                ]);
                builder.sign({ key: testData.TX_SENDER.prv });
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.SIGNED_CONTRACT_WITH_ARGS);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.CONTRACT_ADDRESS);
                tx.outputs[0].value.should.equal('0');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.TX_SENDER.address);
                tx.inputs[0].value.should.equal('0');
            });
            it('a signed self stacking contract call', async () => {
                const builder = initTxBuilder();
                /* Contract call in clarity POX-4
                (define-public (stack-stx (amount-ustx uint)
                                  (pox-addr (tuple (version (buff 1)) (hashbytes (buff 32))))
                                  (start-burn-ht uint)
                                  (lock-period uint)
                                  (signer-sig (optional (buff 65)))
                                  (signer-key (buff 33))
                                  (max-amount uint)
                                  (auth-id uint))
                 */
                builder.functionArgs([
                    { type: 'uint128', val: '400000000' },
                    {
                        type: 'tuple',
                        val: [
                            { key: 'hashbytes', type: 'buffer', val: Buffer.from('some-hash') },
                            { key: 'version', type: 'buffer', val: new bn_js_1.default(1).toBuffer() },
                        ],
                    },
                    { type: 'uint128', val: '52800' },
                    { type: 'uint128', val: '2' },
                    // Nakamoto upgrade new 4 parameters
                    // https://docs.stacks.co/nakamoto-upgrade/signing-and-stacking/stacking-flow#solo-stacker-flow
                    { type: 'optional', val: { type: 'buffer', val: Buffer.from('some-hash') } },
                    { type: 'buffer', val: Buffer.from('some-hash') },
                    { type: 'uint128', val: '340282366920938463463374607431768211455' },
                    { type: 'uint128', val: '123456' },
                ]);
                builder.sign({ key: testData.TX_SENDER.prv });
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.SIGNED_SELF_STACK_CONTRACT_CALL);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.CONTRACT_ADDRESS);
                tx.outputs[0].value.should.equal('0');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.TX_SENDER.address);
                tx.inputs[0].value.should.equal('0');
            });
            it('a signed contract call transaction', async () => {
                const amount = 123;
                const builder = initTxBuilder();
                builder.functionArgs([{ type: 'optional', val: { type: 'int128', val: amount } }]);
                builder.sign({ key: testData.TX_SENDER.prv });
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(txJson.payload.functionArgs, [stringifyCv((0, transactions_1.someCV)((0, transactions_1.intCV)(amount)))]);
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.SIGNED_CONTRACT_CALL);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
            });
            it('a signed serialized contract call transaction', async () => {
                const builder = factory.from(testData.SIGNED_CONTRACT_CALL);
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(txJson.payload.functionArgs, [stringifyCv((0, transactions_1.someCV)((0, transactions_1.intCV)(123)))]);
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.SIGNED_CONTRACT_CALL);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.CONTRACT_ADDRESS);
                tx.outputs[0].value.should.equal('0');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.TX_SENDER.address);
                tx.inputs[0].value.should.equal('0');
            });
            it('a signed serialized self stacking contract call transaction', async () => {
                const builder = factory.from(testData.SIGNED_SELF_STACK_CONTRACT_CALL);
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, testData.CONTRACT_NAME);
                should_1.default.deepEqual(txJson.payload.functionName, testData.CONTRACT_FUNCTION_NAME);
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                // Now stacks-stx self-stacking supports 8 parameters
                // https://docs.stacks.co/nakamoto-upgrade/signing-and-stacking/stacking-flow#solo-stacker-flow
                should_1.default.deepEqual(txJson.payload.functionArgs.length, 8);
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.SIGNED_SELF_STACK_CONTRACT_CALL);
                tx.type.should.equal(sdk_core_1.TransactionType.ContractCall);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.CONTRACT_ADDRESS);
                tx.outputs[0].value.should.equal('0');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.TX_SENDER.address);
                tx.inputs[0].value.should.equal('0');
            });
            it('a multisig transfer transaction', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([{ type: 'optional', val: { type: 'int128', val: '123' } }]);
                builder.sign({ key: testData.prv1 });
                builder.sign({ key: testData.prv2 });
                builder.fromPubKey([testData.pub1, testData.pub2, testData.pub3]);
                builder.numberSignatures(2);
                const tx = await builder.build();
                JSON.stringify(tx.toJson());
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.MULTI_SIG_CONTRACT_CALL);
            });
            describe('ParseCV test', () => {
                it('Optional with out value', () => {
                    const amount = '400000000';
                    const builder = initTxBuilder();
                    builder.functionArgs([
                        { type: 'uint128', val: amount },
                        { type: 'principal', val: testData.ACCOUNT_2.address },
                        { type: 'optional' },
                        {
                            type: 'optional',
                            val: {
                                type: 'tuple',
                                val: [
                                    { key: 'hashbytes', type: 'buffer', val: Buffer.from('some-hash') },
                                    { key: 'version', type: 'buffer', val: new bn_js_1.default(1).toBuffer() },
                                ],
                            },
                        },
                    ]);
                    should_1.default.deepEqual(builder._functionArgs, [
                        (0, transactions_1.uintCV)(amount),
                        (0, transactions_1.standardPrincipalCV)(testData.ACCOUNT_2.address),
                        (0, transactions_1.noneCV)(),
                        (0, transactions_1.someCV)((0, transactions_1.tupleCV)({
                            hashbytes: (0, transactions_1.bufferCV)(Buffer.from('some-hash')),
                            version: (0, transactions_1.bufferCV)(new bn_js_1.default(1).toBuffer()),
                        })),
                    ]);
                });
                it('use ClarityValue', () => {
                    const amount = '400000000';
                    const builder = initTxBuilder();
                    builder.functionArgs([
                        (0, transactions_1.uintCV)(amount),
                        (0, transactions_1.standardPrincipalCV)(testData.ACCOUNT_2.address),
                        (0, transactions_1.noneCV)(),
                        (0, transactions_1.someCV)((0, transactions_1.tupleCV)({
                            hashbytes: (0, transactions_1.bufferCV)(Buffer.from('some-hash')),
                            version: (0, transactions_1.bufferCV)(new bn_js_1.default(1).toBuffer()),
                        })),
                    ]);
                    should_1.default.deepEqual(builder._functionArgs, [
                        (0, transactions_1.uintCV)(amount),
                        (0, transactions_1.standardPrincipalCV)(testData.ACCOUNT_2.address),
                        (0, transactions_1.noneCV)(),
                        (0, transactions_1.someCV)((0, transactions_1.tupleCV)({
                            hashbytes: (0, transactions_1.bufferCV)(Buffer.from('some-hash')),
                            version: (0, transactions_1.bufferCV)(new bn_js_1.default(1).toBuffer()),
                        })),
                    ]);
                });
                it('Buffer as string', () => {
                    const builder = initTxBuilder();
                    builder.functionArgs([
                        { type: 'buffer', val: 'some-hash' },
                        { type: 'buffer', val: '1' },
                    ]);
                    should_1.default.deepEqual(builder._functionArgs, [
                        (0, transactions_1.bufferCV)(Buffer.from('some-hash')),
                        (0, transactions_1.bufferCV)(new bn_js_1.default(1).toBuffer()),
                    ]);
                });
                it('Buffer as number', () => {
                    const builder = initTxBuilder();
                    builder.functionArgs([
                        { type: 'buffer', val: '1' },
                        { type: 'buffer', val: 1 },
                    ]);
                    should_1.default.deepEqual(builder._functionArgs, [
                        (0, transactions_1.bufferCV)(new bn_js_1.default(1).toBuffer()),
                        (0, transactions_1.bufferCV)(new bn_js_1.default(1).toBuffer()),
                    ]);
                });
                it('invalid type', () => {
                    const builder = initTxBuilder();
                    assert_1.default.throws(() => builder.functionArgs([{ type: 'unknow', val: 'any-val' }]), new RegExp('Unexpected Clarity ABI type primitive: "unknow"'));
                });
            });
            describe('should fail', () => {
                it('a contract call with an invalid key', () => {
                    const builder = initTxBuilder();
                    assert_1.default.throws(() => builder.sign({ key: 'invalidKey' }), /Unsupported private key/);
                });
                it('a contract call with an invalid contract address', () => {
                    const builder = initTxBuilder();
                    assert_1.default.throws(() => builder.contractAddress(testData.ACCOUNT_1.address), /Invalid contract address/);
                });
                it('a contract call with an invalid contract name pox-2', () => {
                    const builder = initTxBuilder();
                    assert_1.default.throws(() => builder.contractName('pox-2'), /Only pox-4 and send-many-memo contracts supported/);
                });
                it('a contract call with an invalid contract name pox-3', () => {
                    const builder = initTxBuilder();
                    assert_1.default.throws(() => builder.contractName('pox-3'), /Only pox-4 and send-many-memo contracts supported/);
                });
                it('a contract call with an invalid contract function name', () => {
                    const builder = initTxBuilder();
                    assert_1.default.throws(() => builder.functionName('test-function'), new RegExp('test-function is not supported contract function name'));
                });
            });
        });
    });
});
