// @bitgo/sdk-coin-stx/dist/test/unit/transactionBuilder/fungibleTokenTransferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const statics_1 = require("@bitgo/statics");
const transactions_1 = require("@stacks/transactions");
const should_1 = __importDefault(require("should"));
const network_1 = require("@stacks/network");
const sdk_core_1 = require("@bitgo/sdk-core");
const assert_1 = __importDefault(require("assert"));
const src_1 = require("../../../src");
const testData = __importStar(require("../resources"));
describe('Stacks: Fungible Token Transfer Builder', () => {
    const coinName = 'stx:sbtc';
    const coinNameTest = 'tstx:tsip6dp';
    let bitgo;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, {
            env: 'mock',
        });
        bitgo.initializeTestVars();
        bitgo.safeRegister('tstx:tsip6dp', src_1.Sip10Token.createInstance);
    });
    describe('Sip10 fungible token transfer builder', function () {
        const factory = new src_1.StxLib.TransactionBuilderFactory(statics_1.coins.get(coinNameTest));
        const factoryProd = new src_1.StxLib.TransactionBuilderFactory(statics_1.coins.get(coinName));
        const initTxBuilder = () => {
            const txBuilder = factory.getFungibleTokenTransferBuilder();
            txBuilder.fee({ fee: '180' });
            txBuilder.nonce(0);
            txBuilder.contractAddress(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.CONTRACT_ADDRESS);
            txBuilder.contractName(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.CONTRACT_NAME);
            txBuilder.functionName(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.FUNCTION_NAME);
            txBuilder.tokenName(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.TOKEN_NAME);
            return txBuilder;
        };
        describe('fungible token builder environment', function () {
            it('should select the right network', function () {
                should_1.default.equal(factory.getFungibleTokenTransferBuilder().coinName(), 'tstx:tsip6dp');
                should_1.default.equal(factoryProd.getFungibleTokenTransferBuilder().coinName(), 'stx:sbtc');
                // used type any to access protected properties
                const txBuilder = factory.getFungibleTokenTransferBuilder();
                const txBuilderProd = factoryProd.getFungibleTokenTransferBuilder();
                txBuilder._network.should.deepEqual(new network_1.StacksTestnet());
                txBuilderProd._network.should.deepEqual(new network_1.StacksMainnet());
            });
        });
        describe('should build', function () {
            it('an unsigned fungible token transfer transaction', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([
                    (0, transactions_1.uintCV)('10000'),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SENDER_ADDRESS),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.RECEIVER_ADDRESS),
                    (0, transactions_1.someCV)((0, transactions_1.bufferCVFromString)('1')),
                ]);
                builder.fromPubKey(testData.TX_SENDER.pub);
                builder.numberSignatures(1);
                const tx = await builder.build();
                const stacksTxn = tx;
                should_1.default.deepEqual(stacksTxn.stxTransaction.postConditions.values.length, 1);
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, 'tsip6dp-token');
                should_1.default.deepEqual(txJson.payload.functionName, 'transfer');
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.UNSIGNED_SINGLE_SIG_TX);
                tx.type.should.equal(sdk_core_1.TransactionType.Send);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.RECEIVER_ADDRESS);
                tx.outputs[0].value.should.equal('10000');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SENDER_ADDRESS);
                tx.inputs[0].value.should.equal('10000');
            });
            it('an unsigned fungible token transfer transaction without memo', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([
                    (0, transactions_1.uintCV)('10000'),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SENDER_ADDRESS),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.RECEIVER_ADDRESS),
                ]);
                builder.fromPubKey(testData.TX_SENDER.pub);
                builder.numberSignatures(1);
                const tx = await builder.build();
                const txJson = tx.toJson();
                should_1.default.deepEqual(txJson.payload.contractAddress, testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.CONTRACT_ADDRESS);
                should_1.default.deepEqual(txJson.payload.contractName, 'tsip6dp-token');
                should_1.default.deepEqual(txJson.payload.functionName, 'transfer');
                should_1.default.deepEqual(txJson.nonce, 0);
                should_1.default.deepEqual(txJson.fee.toString(), '180');
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.UNSIGNED_SINGLE_SIG_TX_WITHOUT_MEMO);
                tx.type.should.equal(sdk_core_1.TransactionType.Send);
                tx.outputs.length.should.equal(1);
                tx.outputs[0].address.should.equal(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.RECEIVER_ADDRESS);
                tx.outputs[0].value.should.equal('10000');
                tx.inputs.length.should.equal(1);
                tx.inputs[0].address.should.equal(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SENDER_ADDRESS);
                tx.inputs[0].value.should.equal('10000');
            });
            it('a multisig fungible token transfer transaction with memo', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([
                    (0, transactions_1.uintCV)('10000'),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SENDER_ADDRESS),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.RECEIVER_ADDRESS),
                    (0, transactions_1.someCV)((0, transactions_1.bufferCVFromString)('1')),
                ]);
                builder.sign({ key: testData.prv1 });
                builder.sign({ key: testData.prv2 });
                builder.fromPubKey([testData.pub1, testData.pub2, testData.pub3]);
                builder.numberSignatures(2);
                const tx = await builder.build();
                JSON.stringify(tx.toJson());
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SIGNED_MULTI_SIG_TX);
            });
            it('a multisig fungible token transfer transaction without memo', async () => {
                const builder = initTxBuilder();
                builder.functionArgs([
                    (0, transactions_1.uintCV)('10000'),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SENDER_ADDRESS),
                    (0, transactions_1.standardPrincipalCV)(testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.RECEIVER_ADDRESS),
                    (0, transactions_1.noneCV)(),
                ]);
                builder.sign({ key: testData.prv1 });
                builder.sign({ key: testData.prv2 });
                builder.fromPubKey([testData.pub1, testData.pub2, testData.pub3]);
                builder.numberSignatures(2);
                const tx = await builder.build();
                JSON.stringify(tx.toJson());
                should_1.default.deepEqual(tx.toBroadcastFormat(), testData.FUNGIBLE_TOKEN_TRANSFER_CONSTANTS.SIGNED_MULTI_SIG_TX_WITHOUT_MEMO);
            });
        });
        describe('should fail', () => {
            it('a fungible token transfer builder with an invalid key', () => {
                const builder = initTxBuilder();
                assert_1.default.throws(() => builder.sign({ key: 'invalidKey' }), /Unsupported private key/);
            });
            it('a fungible token transfer builder with an invalid contract address', () => {
                const builder = initTxBuilder();
                assert_1.default.throws(() => builder.contractAddress('invalidContractAddress'), /Invalid address/);
            });
            it('a fungible token transfer builder with an invalid contract function name', () => {
                const builder = initTxBuilder();
                assert_1.default.throws(() => builder.functionName('test-function'), new RegExp('test-function is not supported contract function name'));
            });
        });
    });
});
