// @bitgo/sdk-coin-sgb/dist/src/sgb.js
"use strict";
/***
 * @developerReferences
 * - Developer Hub: https://dev.flare.network/
 * - Doc: https://docs.flare.network/user/wallets/
 * @coinFullName Songbird
 * @coinWebsite https://songbird-explorer.flare.network
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sgb = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class Sgb extends abstract_eth_1.AbstractEthLikeNewCoins {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
    }
    static createInstance(bitgo, staticsCoin) {
        return new Sgb(bitgo, staticsCoin);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    async buildUnsignedSweepTxnTSS(params) {
        return this.buildUnsignedSweepTxnMPCv2(params);
    }
    /**
     * Make a query to Songbird explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @param {string} apiKey optional API key to use instead of the one from the environment
     * @returns {Promise<Object>} response from Songbird explorer
     */
    async recoveryBlockchainExplorerQuery(query, apiKey) {
        const apiToken = apiKey || sdk_core_1.common.Environments[this.bitgo.getEnv()].sgbExplorerApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].sgbExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
}
exports.Sgb = Sgb;
