// ic0/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpAgent = exports.mockCanister = exports.devCanister = exports.agentCanister = exports.replica = exports.local = exports.ic = void 0;
const agent_1 = require("@dfinity/agent");
Object.defineProperty(exports, "HttpAgent", { enumerable: true, get: function () { return agent_1.HttpAgent; } });
const agentCanister_1 = require("./canister/agentCanister");
Object.defineProperty(exports, "agentCanister", { enumerable: true, get: function () { return agentCanister_1.agentCanister; } });
const devCanister_1 = require("./canister/devCanister");
Object.defineProperty(exports, "devCanister", { enumerable: true, get: function () { return devCanister_1.devCanister; } });
const mockCanister_1 = require("./canister/mockCanister");
Object.defineProperty(exports, "mockCanister", { enumerable: true, get: function () { return mockCanister_1.mockCanister; } });
const replica_1 = require("./replica");
Object.defineProperty(exports, "replica", { enumerable: true, get: function () { return replica_1.replica; } });
const ic = (0, replica_1.deferredReplica)('https://icp-api.io');
exports.ic = ic;
const local = (0, replica_1.deferredReplica)('http://localhost:4943', { local: true });
exports.local = local;
// Configure exports for TS, CommonJS, and ESM
const defaultExport = ic;
Object.assign(defaultExport, {
    default: defaultExport,
    ic,
    local,
    replica: replica_1.replica,
    agentCanister: agentCanister_1.agentCanister,
    devCanister: devCanister_1.devCanister,
    mockCanister: mockCanister_1.mockCanister,
    HttpAgent: agent_1.HttpAgent,
    __esModule: true, // Recognize as ES Module
});
module.exports = defaultExport;
exports.default = defaultExport;
