// @bitgo/abstract-utxo/dist/cjs/test/unit/transaction/outputDifference.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const outputDifference_1 = require("../../../src/transaction/outputDifference");
describe('outputDifference', function () {
    function output(script, value, optional) {
        const scriptBuffer = Buffer.from(script, 'hex');
        if (scriptBuffer.toString('hex') !== script) {
            throw new Error('invalid script');
        }
        return {
            script: Buffer.from(script, 'hex'),
            value: value === 'max' ? 'max' : BigInt(value),
            ...(optional !== undefined ? { optional } : {}),
        };
    }
    function expectedOutput(script, value, optional) {
        return output(script, value, optional);
    }
    const a = output('aa', 1);
    const a2 = output('aa', 2);
    const aMax = output('aa', 'max');
    const b = output('bb', 1);
    const c = output('cc', 1);
    describe('equalOutput', function () {
        it('has expected result', function () {
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(a, a), true);
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(a, a2), false);
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(a, b), false);
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(aMax, b), false);
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(aMax, a), true);
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(a, aMax), true);
            // this one does not appear in practice but is a valid comparison
            assert_1.default.deepStrictEqual((0, outputDifference_1.matchingOutput)(aMax, aMax), true);
        });
    });
    describe('outputDifference', function () {
        it('has expected result', function () {
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([], []), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a], []), [a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([aMax], []), [aMax]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([], [a]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([], [aMax]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a], [a]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a], [aMax]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([aMax], [a]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a, a], [a]), [a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a, a], [aMax]), [a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a, a, a], [a]), [a, a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a, b, c], [a, b]), [c]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a, b, c], [aMax, b]), [c]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a, b, c, a], [a, b]), [c, a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a], [a2]), [a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.outputDifference)([a2], [a]), [a2]);
        });
    });
    describe('getMissingOutputs', function () {
        it('returns missing non-optional outputs', function () {
            const aOptional = expectedOutput('aa', 1, true);
            const bOptional = expectedOutput('bb', 1, true);
            // No expected outputs means no missing outputs
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([a], []), []);
            // Missing required output is returned
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([], [a]), [a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([b], [a]), [a]);
            // Missing optional output is filtered out
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([], [aOptional]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([b], [aOptional]), []);
            // Mix of optional and required: only required missing outputs returned
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([], [a, bOptional]), [a]);
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([], [aOptional, b]), [b]);
            // Present outputs are not returned regardless of optional flag
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([a], [a]), []);
            assert_1.default.deepStrictEqual((0, outputDifference_1.getMissingOutputs)([a], [aOptional]), []);
        });
    });
    describe('outputDifferencesWithExpected', function () {
        function test(outputs, recipients, expected) {
            const result = (0, outputDifference_1.outputDifferencesWithExpected)(outputs, recipients);
            assert_1.default.deepStrictEqual(result, {
                explicitOutputs: expected.explicit,
                implicitOutputs: expected.implicit,
                missingOutputs: expected.missing,
            });
        }
        it('has expected result', function () {
            test([a], [], { missing: [], explicit: [], implicit: [a] });
            test([], [a], { missing: [a], explicit: [], implicit: [] });
            test([a], [a], { missing: [], explicit: [a], implicit: [] });
            test([a], [a2], { missing: [a2], explicit: [], implicit: [a] });
            test([b], [a], { missing: [a], explicit: [], implicit: [b] });
            test([a, a], [a], { missing: [], explicit: [a], implicit: [a] });
            test([a, b], [a], { missing: [], explicit: [a], implicit: [b] });
        });
        it('handles optional expected outputs', function () {
            const aOptional = expectedOutput('aa', 1, true);
            const bOptional = expectedOutput('bb', 1, true);
            // Missing optional output is not reported as missing
            test([], [aOptional], { missing: [], explicit: [], implicit: [] });
            test([b], [aOptional], { missing: [], explicit: [], implicit: [b] });
            // Present optional output is still explicit
            test([a], [aOptional], { missing: [], explicit: [a], implicit: [] });
            // Mix of required and optional outputs
            test([], [a, bOptional], { missing: [a], explicit: [], implicit: [] });
            test([a], [a, bOptional], { missing: [], explicit: [a], implicit: [] });
            test([a, b], [a, bOptional], { missing: [], explicit: [a, b], implicit: [] });
            // Multiple optional outputs
            test([], [aOptional, bOptional], { missing: [], explicit: [], implicit: [] });
            test([a], [aOptional, bOptional], { missing: [], explicit: [a], implicit: [] });
            test([a, b], [aOptional, bOptional], { missing: [], explicit: [a, b], implicit: [] });
        });
    });
});
