// @bitgo/sdk-coin-canton/dist/src/lib/oneStepPreApprovalBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneStepPreApprovalBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class OneStepPreApprovalBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        this.setTransactionType();
    }
    get transactionType() {
        return sdk_core_1.TransactionType.OneStepPreApproval;
    }
    setTransactionType() {
        this.transaction.transactionType = sdk_core_1.TransactionType.OneStepPreApproval;
    }
    setTransaction(transaction) {
        this.transaction.prepareCommand = transaction;
    }
    /** @inheritDoc */
    addSignature(publicKey, signature) {
        if (!this.transaction) {
            throw new sdk_core_1.InvalidTransactionError('transaction is empty!');
        }
        this._signatures.push({ publicKey, signature });
        const pubKeyBase64 = utils_1.default.getBase64FromHex(publicKey.pub);
        this.transaction.signerFingerprint = utils_1.default.getAddressFromPublicKey(pubKeyBase64);
        this.transaction.signatures = signature.toString('base64');
    }
    /**
     * Sets the unique id for the 1-step enablement
     * Also sets the _id of the transaction
     *
     * @param id - A uuid
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    commandId(id) {
        if (!id.trim()) {
            throw new Error('commandId must be a non-empty string');
        }
        this._commandId = id.trim();
        // also set the transaction _id
        this.transaction.id = id.trim();
        return this;
    }
    /**
     * Sets the receiver for the 1-step enablement
     *
     * @param id - the receiver party id (address)
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    receiverPartyId(id) {
        if (!id.trim()) {
            throw new Error('receiverPartyId must be a non-empty string');
        }
        this._receiverPartyId = id.trim();
        return this;
    }
    /**
     * Sets the optional token field if present, used for canton token preApproval setup
     * @param name - the bitgo name of the token
     * @returns The current builder for chaining
     * @throws Error if name is invalid
     */
    tokenName(name) {
        if (!name || !name.trim()) {
            throw new Error('token name must be a non-empty string');
        }
        this._tokenName = name.trim();
        return this;
    }
    /**
     * Builds and returns the CantonOneStepEnablementRequest object from the builder's internal state.
     *
     * This method performs validation before constructing the object. If required fields are
     * missing or invalid, it throws an error.
     *
     * @returns {CantonOneStepEnablementRequest} - A fully constructed and validated request object for 1-step enablement.
     * @throws {Error} If any required field is missing or fails validation.
     */
    toRequestObject() {
        this.validate();
        return {
            commandId: this._commandId,
            receiverId: this._receiverPartyId,
            verboseHashing: false,
            actAs: [this._receiverPartyId],
            readAs: [],
            tokenName: this._tokenName,
        };
    }
    /**
     * Validates the internal state of the builder before building the request object.
     *
     * @private
     * @throws {Error} If any required field is missing or invalid.
     */
    validate() {
        if (!this._commandId)
            throw new Error('commandId is missing');
        if (!this._receiverPartyId)
            throw new Error('receiver partyId is missing');
    }
}
exports.OneStepPreApprovalBuilder = OneStepPreApprovalBuilder;
