// @bitgo/sdk-coin-canton/dist/src/lib/transactionBuilderFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const oneStepPreApprovalBuilder_1 = require("./oneStepPreApprovalBuilder");
const transferAcceptanceBuilder_1 = require("./transferAcceptanceBuilder");
const transferAcknowledgeBuilder_1 = require("./transferAcknowledgeBuilder");
const transferBuilder_1 = require("./transferBuilder");
const transferOfferWithdrawnBuilder_1 = require("./transferOfferWithdrawnBuilder");
const transferRejectionBuilder_1 = require("./transferRejectionBuilder");
const transaction_1 = require("./transaction/transaction");
const walletInitBuilder_1 = require("./walletInitBuilder");
const walletInitTransaction_1 = require("./walletInitialization/walletInitTransaction");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    from(raw) {
        try {
            const tx = new walletInitTransaction_1.WalletInitTransaction(this._coinConfig);
            tx.fromRawTransaction(raw);
            return this.getWalletInitializationBuilder(tx);
        }
        catch {
            const tx = new transaction_1.Transaction(this._coinConfig);
            tx.fromRawTransaction(raw);
            switch (tx.type) {
                case sdk_core_1.TransactionType.OneStepPreApproval: {
                    return this.getOneStepPreapprovalBuilder(tx);
                }
                case sdk_core_1.TransactionType.Send: {
                    return this.getTransferBuilder(tx);
                }
                case sdk_core_1.TransactionType.TransferAccept: {
                    return this.getTransferAcceptanceBuilder(tx);
                }
                case sdk_core_1.TransactionType.TransferAcknowledge: {
                    return this.getTransferAcknowledgeBuilder(tx);
                }
                case sdk_core_1.TransactionType.TransferOfferWithdrawn: {
                    return this.getTransferOfferWithdrawnBuilder(tx);
                }
                case sdk_core_1.TransactionType.TransferReject: {
                    return this.getTransferRejectBuilder(tx);
                }
                default: {
                    throw new sdk_core_1.InvalidTransactionError('unsupported transaction');
                }
            }
        }
    }
    getOneStepPreapprovalBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new oneStepPreApprovalBuilder_1.OneStepPreApprovalBuilder(this._coinConfig));
    }
    getTransferAcceptanceBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferAcceptanceBuilder_1.TransferAcceptanceBuilder(this._coinConfig));
    }
    getTransferAcknowledgeBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferAcknowledgeBuilder_1.TransferAcknowledgeBuilder(this._coinConfig));
    }
    getTransferOfferWithdrawnBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferOfferWithdrawnBuilder_1.TransferOfferWithdrawnBuilder(this._coinConfig));
    }
    getTransferRejectBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferRejectionBuilder_1.TransferRejectionBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getTransferBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getWalletInitializationBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new walletInitBuilder_1.WalletInitBuilder(this._coinConfig));
    }
    static initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
