// @bitgo/sdk-coin-canton/dist/src/lib/transferAcknowledgeBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferAcknowledgeBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
class TransferAcknowledgeBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        this.setTransactionType();
    }
    get transactionType() {
        return sdk_core_1.TransactionType.TransferAcknowledge;
    }
    setTransactionType() {
        this.transaction.transactionType = sdk_core_1.TransactionType.TransferAcknowledge;
    }
    setTransaction(transaction) {
        throw new Error('Not implemented!');
    }
    /** @inheritDoc */
    addSignature(publicKey, signature) {
        throw new Error('Not implemented!');
    }
    /**
     * Sets the contract id the receiver needs to accept/reject
     * @param id - canton contract id
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    contractId(id) {
        if (!id || !id.trim()) {
            throw new Error('contractId must be a non-empty string');
        }
        this._contractId = id.trim();
        return this;
    }
    /**
     * Sets the sender party id
     * @param id - sender party id (address)
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    senderPartyId(id) {
        if (!id || !id.trim()) {
            throw new Error('partyId must be a non-empty string');
        }
        this._senderPartyId = id.trim();
        return this;
    }
    /**
     * Sets the amount to accept or reject
     * @param amount - incoming deposit amount
     * @returns The current builder instance for chaining.
     * @throws Error if amount <= 0
     */
    amount(amount) {
        if (isNaN(amount) || amount <= 0) {
            throw new Error('amount must be positive number');
        }
        this._amount = amount;
        return this;
    }
    /**
     * Sets the incoming txn id (updateId of the ledger update)
     * @param id - ledger update id
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    updateId(id) {
        if (!id || !id.trim()) {
            throw new Error('updateId must be a non-empty string');
        }
        this._updateId = id.trim();
        // also set the transaction id
        this.transaction.id = id.trim();
        return this;
    }
    /**
     * Sets the deposit txn expiry
     * @param expiry - expiry epoch
     * @returns The current builder instance for chaining.
     * @throws Error if epoch is invalid
     */
    expiryEpoch(expiry) {
        if (isNaN(expiry)) {
            throw new Error('invalid expiry');
        }
        this._expiryEpoch = expiry;
        return this;
    }
    /**
     * Builds and returns the TransferAcknowledge object from the builder's internal state.
     *
     * This method performs validation before constructing the object. If required fields are
     * missing or invalid, it throws an error.
     *
     * @returns {TransferAcknowledge} - A fully constructed and validated request object for transfer acknowledge.
     * @throws {Error} If any required field is missing or fails validation.
     */
    toRequestObject() {
        this.validate();
        return {
            contractId: this._contractId,
            senderPartyId: this._senderPartyId,
            updateId: this._updateId,
            amount: this._amount,
            expiryEpoch: this._expiryEpoch,
        };
    }
    /**
     * Validates the internal state of the builder before building the request object.
     *
     * @private
     * @throws {Error} If any required field is missing or invalid.
     */
    validate() {
        if (!this._contractId)
            throw new Error('contractId is missing');
        if (!this._updateId)
            throw new Error('updateId is missing');
        if (!this._senderPartyId)
            throw new Error('sender partyId is missing');
        if (!this._amount)
            throw new Error('amount is missing');
        if (!this._expiryEpoch)
            throw new Error('expiry is missing');
    }
}
exports.TransferAcknowledgeBuilder = TransferAcknowledgeBuilder;
