// @bitgo/sdk-coin-canton/dist/src/lib/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class TransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._sendOneStep = false;
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        this.setTransactionType();
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    setTransactionType() {
        this.transaction.transactionType = sdk_core_1.TransactionType.Send;
    }
    setTransaction(transaction) {
        this.transaction.prepareCommand = transaction;
    }
    /** @inheritDoc */
    addSignature(publicKey, signature) {
        if (!this.transaction) {
            throw new sdk_core_1.InvalidTransactionError('transaction is empty!');
        }
        this._signatures.push({ publicKey, signature });
        const pubKeyBase64 = utils_1.default.getBase64FromHex(publicKey.pub);
        this.transaction.signerFingerprint = utils_1.default.getAddressFromPublicKey(pubKeyBase64);
        this.transaction.signatures = signature.toString('base64');
    }
    /**
     * Sets the unique id for the transfer
     * Also sets the _id of the transaction
     *
     * @param id - A uuid
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    commandId(id) {
        if (!id || !id.trim()) {
            throw new Error('commandId must be a non-empty string');
        }
        this._commandId = id.trim();
        // also set the transaction _id
        this.transaction.id = id.trim();
        return this;
    }
    /**
     * Sets the sender party id for the transfer
     * @param id - sender address (party id)
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    senderId(id) {
        if (!id || !id.trim()) {
            throw new Error('senderId must be a non-empty string');
        }
        this._senderId = id.trim();
        return this;
    }
    /**
     * Sets the receiver party id for the transfer
     * @param id - receiver address (party id)
     * @returns The current builder instance for chaining.
     * @throws Error if id is empty.
     */
    receiverId(id) {
        if (!id || !id.trim()) {
            throw new Error('receiverId must be a non-empty string');
        }
        this._receiverId = id.trim();
        return this;
    }
    /**
     * Sets the transfer amount
     * @param amount - transfer amount
     * @returns The current builder instance for chaining.
     * @throws Error if amount not present or negative
     */
    amount(amount) {
        if (!amount || amount < 0) {
            throw new Error('amount must be a positive number');
        }
        this._amount = amount;
        return this;
    }
    /**
     * Sets the 1-step enablement flag to send via 1-step, works only if recipient
     * enabled the 1-step, defaults to `false`
     * @param flag boolean value
     * @returns The current builder for chaining
     */
    sendOneStep(flag) {
        this._sendOneStep = flag;
        return this;
    }
    /**
     * Sets the transfer expiry
     * @param epoch - the expiry for 2-step transfer, defaults to 90 days and
     * not applicable if sending via 1-step
     * @returns The current builder for chaining
     * @throws Error if the expiry value is invalid
     */
    expiryEpoch(epoch) {
        if (!epoch || epoch < 0) {
            throw new Error('epoch must be a positive number');
        }
        this._expiryEpoch = epoch;
        return this;
    }
    /**
     * Sets the optional memoId if present
     * @param id - memoId of the recipient
     * @returns The current builder for chaining
     * @throws Error if the memoId value is invalid
     */
    memoId(id) {
        if (!id || !id.trim()) {
            throw new Error('memoId must be a non-empty string');
        }
        this._memoId = id.trim();
        return this;
    }
    /**
     * Sets the optional token field if present, used for canton token transaction
     * @param name - the bitgo name of the token
     * @returns The current builder for chaining
     * @throws Error if name is invalid
     */
    tokenName(name) {
        if (!name || !name.trim()) {
            throw new Error('token name must be a non-empty string');
        }
        this._tokenName = name.trim();
        return this;
    }
    /**
     * Get the canton transfer request object
     * @returns CantonTransferRequest
     * @throws Error if any required params are missing
     */
    toRequestObject() {
        this.validate();
        const data = {
            commandId: this._commandId,
            senderPartyId: this._senderId,
            receiverPartyId: this._receiverId,
            amount: this._amount,
            expiryEpoch: this._expiryEpoch,
            sendViaOneStep: this._sendOneStep,
        };
        if (this._memoId) {
            data.memoId = this._memoId;
        }
        if (this._tokenName) {
            data.tokenName = this._tokenName;
        }
        return data;
    }
    /**
     * Method to validate the required fields
     * @throws Error if required fields are not set
     * @private
     */
    validate() {
        if (!this._commandId)
            throw new Error('commandId is missing');
        if (!this._senderId)
            throw new Error('senderId is missing');
        if (!this._receiverId)
            throw new Error('receiverId is missing');
        if (!this._amount)
            throw new Error('amount is missing');
        if (!this._expiryEpoch)
            throw new Error('expiryEpoch is missing');
    }
}
exports.TransferBuilder = TransferBuilder;
