// @bitgo/sdk-coin-canton/dist/src/lib/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const keyPair_1 = require("./keyPair");
const utils_1 = __importDefault(require("./utils"));
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor() {
        super(...arguments);
        this._signatures = [];
    }
    initBuilder(tx) {
        this._transaction = tx;
    }
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    /** @inheritdoc */
    signImplementation(key) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    async buildImplementation() {
        throw new Error('Method not implemented.');
    }
    // region Validators
    /** @inheritdoc */
    validateAddress(address, addressFormat) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address ' + address.address);
        }
    }
    /** @inheritdoc */
    validateKey(key) {
        let keyPair;
        try {
            keyPair = new keyPair_1.KeyPair({ prv: key.key });
        }
        catch {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
        if (!keyPair.getKeys().prv) {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
    }
    /** @inheritdoc */
    async validateRawTransaction(rawTransaction) {
        if (!rawTransaction || !this._transaction.prepareCommand) {
            throw new sdk_core_1.BuildTransactionError('invalid raw transaction');
        }
        const localHash = await utils_1.default.computeHashFromPrepareSubmissionResponse(rawTransaction);
        if (localHash !== this._transaction.prepareCommand.preparedTransactionHash) {
            throw new sdk_core_1.BuildTransactionError('invalid raw transaction, hash not matching');
        }
    }
    /** @inheritdoc */
    async validateTransaction(transaction) {
        if (!transaction?.prepareCommand?.preparedTransaction) {
            return;
        }
        const localHash = await utils_1.default.computeHashFromPrepareSubmissionResponse(transaction.prepareCommand.preparedTransaction);
        if (localHash !== transaction.prepareCommand.preparedTransactionHash) {
            throw new sdk_core_1.BuildTransactionError('invalid transaction');
        }
    }
    /** @inheritdoc */
    validateValue(value) {
        if (value.isLessThan(0)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be less than zero');
        }
    }
}
exports.TransactionBuilder = TransactionBuilder;
