// @bitgo/sdk-coin-canton/dist/src/lib/walletInitBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletInitBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const keyPair_1 = require("./keyPair");
const utils_1 = __importDefault(require("./utils"));
const walletInitTransaction_1 = require("./walletInitialization/walletInitTransaction");
const constant_1 = require("./constant");
class WalletInitBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._signatures = [];
        this._localParticipantObservationOnly = false;
        this._otherConfirmingParticipantUids = [];
        this._confirmationThreshold = 1;
        this._observingParticipantUids = [];
        this._transaction = new walletInitTransaction_1.WalletInitTransaction(_coinConfig);
    }
    initBuilder(tx) {
        this._transaction = tx;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.WalletInitialization;
    }
    buildImplementation() {
        throw new Error('Not implemented');
    }
    fromImplementation(rawTransaction, isFirstSigner) {
        throw new Error('Not implemented');
    }
    signImplementation(key) {
        throw new Error('Not implemented');
    }
    get transaction() {
        return this._transaction;
    }
    set transaction(transaction) {
        this._transaction.preparedParty = transaction;
    }
    validateAddress(address, addressFormat) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address ' + address.address);
        }
    }
    validateKey(key) {
        let keyPair;
        try {
            keyPair = new keyPair_1.KeyPair({ prv: key.key });
        }
        catch {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
        if (!keyPair.getKeys().prv) {
            throw new sdk_core_1.BuildTransactionError('Invalid key');
        }
    }
    validateRawTransaction(rawTransaction) {
        if (!rawTransaction || !this._transaction.preparedParty) {
            throw new sdk_core_1.BuildTransactionError('invalid raw transaction');
        }
        const localHash = utils_1.default.computeHashFromCreatePartyResponse(rawTransaction);
        if (localHash !== this._transaction.preparedParty.multiHash) {
            throw new sdk_core_1.BuildTransactionError('invalid raw transaction, hash not matching');
        }
    }
    validateTransaction(transaction) {
        if (!transaction?.preparedParty) {
            return;
        }
        const localHash = utils_1.default.computeHashFromCreatePartyResponse(transaction.preparedParty.topologyTransactions);
        if (localHash !== this._transaction.preparedParty.multiHash) {
            throw new sdk_core_1.BuildTransactionError('invalid transaction');
        }
    }
    // not implemented because wallet initialization doesn't require any value
    validateValue(value) {
        throw new Error('Not implemented');
    }
    /** @inheritDoc */
    addSignature(publicKey, signature) {
        if (!this.transaction) {
            throw new sdk_core_1.InvalidTransactionError('transaction is empty!');
        }
        this._signatures.push({ publicKey, signature });
        this.transaction.signatures = signature.toString('base64');
    }
    /**
     * Sets the public key used for signing.
     *
     * @param key - The public key of the wallet.
     * @returns The current builder instance for chaining.
     * @throws Error if key is not a valid public key
     */
    publicKey(key) {
        utils_1.default.isValidPublicKey(key);
        this._publicKey = {
            format: constant_1.PUBLIC_KEY_FORMAT,
            keyData: key,
            keySpec: constant_1.PUBLIC_KEY_SPEC,
        };
        return this;
    }
    /**
     * Sets the party hint (alias or name) used during wallet initialization.
     *
     * @param hint - A 5-character non-empty string representing the party hint.
     * @returns The current builder instance for chaining.
     * @throws Error if the hint is empty or more than 5 characters long.
     */
    partyHint(hint) {
        const trimmedHint = hint.trim();
        if (!trimmedHint) {
            throw new Error('partyHint cannot be empty');
        }
        if (trimmedHint.length > 5) {
            throw new Error('partyHint must be less than 6 characters long');
        }
        this._partyHint = trimmedHint;
        return this;
    }
    /**
     * Sets whether the local participant should be added in observation-only mode.
     *
     * @param flag - Boolean flag indicating observation-only status.
     * @returns The current builder instance for chaining.
     */
    localParticipantObservationOnly(flag) {
        this._localParticipantObservationOnly = flag;
        return this;
    }
    /**
     * Adds a confirming participant UID to the list of other confirming participants.
     *
     * @param uid - A non-empty string UID of another confirming participant.
     * @returns The current builder instance for chaining.
     * @throws Error if the UID is empty.
     */
    otherConfirmingParticipantUid(uid) {
        const trimmedUid = uid.trim();
        if (!trimmedUid) {
            throw new Error('otherConfirmingParticipantUid cannot be empty');
        }
        if (!this._otherConfirmingParticipantUids) {
            this._otherConfirmingParticipantUids = [];
        }
        this._otherConfirmingParticipantUids.push(trimmedUid);
        return this;
    }
    /**
     * Sets the confirmation threshold for the wallet initialization.
     *
     * @param threshold - A positive integer indicating how many confirmations are required.
     * @returns The current builder instance for chaining.
     * @throws Error if the threshold is not a positive integer.
     */
    confirmationThreshold(threshold) {
        if (!Number.isInteger(threshold) || threshold <= 0) {
            throw new Error('confirmationThreshold must be a positive integer');
        }
        this._confirmationThreshold = threshold;
        return this;
    }
    /**
     * Adds an observing participant UID to the list of observers.
     *
     * @param uid - A non-empty string UID of the observing participant.
     * @returns The current builder instance for chaining.
     * @throws Error if the UID is empty.
     */
    observingParticipantUid(uid) {
        const trimmedUid = uid.trim();
        if (!trimmedUid) {
            throw new Error('observingParticipantUid cannot be empty');
        }
        if (!this._observingParticipantUids) {
            this._observingParticipantUids = [];
        }
        this._observingParticipantUids.push(trimmedUid);
        return this;
    }
    /**
     * Builds and returns the WalletInitRequest object from the builder's internal state.
     *
     * This method performs validation before constructing the object. If required fields are
     * missing or invalid, it throws an error.
     *
     * @returns {WalletInitRequest} - A fully constructed and validated request object for wallet initialization.
     * @throws {Error} If any required field is missing or fails validation.
     */
    toRequestObject() {
        this.validate();
        return {
            publicKey: this._publicKey,
            partyHint: this._partyHint,
            localParticipantObservationOnly: this._localParticipantObservationOnly,
            otherConfirmingParticipantUids: this._otherConfirmingParticipantUids,
            confirmationThreshold: this._confirmationThreshold,
            observingParticipantUids: this._observingParticipantUids,
        };
    }
    /**
     * Validates the internal state of the builder before building the request object.
     *
     * Checks for:
     * - Presence of a synchronizer ID.
     * - `partyHint` must not be empty and must be at most 5 characters long.
     * - Public key must include `keyData`, `format`, and `keySpec`.
     * - Confirmation threshold must be a positive integer.
     * - Confirming and observing participant UID lists must be arrays.
     *
     * @private
     * @throws {Error} If any required field is missing or invalid.
     */
    validate() {
        if (!this._partyHint || this._partyHint.length > 5)
            throw new Error('Invalid partyHint');
        if (!this._publicKey || !this._publicKey.keyData || !this._publicKey.format || !this._publicKey.keySpec) {
            throw new Error('Invalid publicKey');
        }
        if (!Number.isInteger(this._confirmationThreshold) || this._confirmationThreshold <= 0) {
            throw new Error('Invalid confirmationThreshold');
        }
        if (!Array.isArray(this._otherConfirmingParticipantUids)) {
            throw new Error('otherConfirmingParticipantUids must be an array');
        }
        if (!Array.isArray(this._observingParticipantUids)) {
            throw new Error('observingParticipantUids must be an array');
        }
    }
}
exports.WalletInitBuilder = WalletInitBuilder;
