// @bitgo/sdk-core/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.common = exports.verifyMPCWalletAddress = exports.verifyEddsaTssWalletAddress = exports.EcdsaMPCv2Utils = exports.EcdsaUtils = exports.EddsaUtils = exports.coins = exports.accountLib = void 0;
__exportStar(require("./account-lib"), exports);
exports.accountLib = __importStar(require("./account-lib"));
__exportStar(require("./api"), exports);
__exportStar(require("./bitgo"), exports);
__exportStar(require("./bitgojsError"), exports);
exports.coins = __importStar(require("./coins"));
const eddsa_1 = require("./bitgo/utils/tss/eddsa/eddsa");
Object.defineProperty(exports, "EddsaUtils", { enumerable: true, get: function () { return eddsa_1.EddsaUtils; } });
const ecdsa_1 = require("./bitgo/utils/tss/ecdsa/ecdsa");
Object.defineProperty(exports, "EcdsaUtils", { enumerable: true, get: function () { return ecdsa_1.EcdsaUtils; } });
const ecdsaMPCv2_1 = require("./bitgo/utils/tss/ecdsa/ecdsaMPCv2");
Object.defineProperty(exports, "EcdsaMPCv2Utils", { enumerable: true, get: function () { return ecdsaMPCv2_1.EcdsaMPCv2Utils; } });
var addressVerification_1 = require("./bitgo/utils/tss/addressVerification");
Object.defineProperty(exports, "verifyEddsaTssWalletAddress", { enumerable: true, get: function () { return addressVerification_1.verifyEddsaTssWalletAddress; } });
Object.defineProperty(exports, "verifyMPCWalletAddress", { enumerable: true, get: function () { return addressVerification_1.verifyMPCWalletAddress; } });
const common = __importStar(require("./common"));
exports.common = common;
__exportStar(require("./units"), exports);
