// @bitgo/public-types/dist/src/schema/webhook/addWalletWebhookRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddWalletWebhookRequestBodyC = exports.AddWalletWebhookRequestBody = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const optional_1 = require("../../utils/optional");
const transactionRequestState_1 = require("../transactionRequest/transactionRequestState");
const transactionState_1 = require("../transactionRequest/transactionState");
const webhookType_1 = require("./webhookType");
exports.AddWalletWebhookRequestBody = {
    type: webhookType_1.WalletWebhookType,
    url: io_ts_types_1.NonEmptyString,
    label: (0, optional_1.Optional)(t.string),
    numConfirmations: t.union([t.number, io_ts_types_1.NumberFromString, t.undefined]),
    allToken: (0, optional_1.Optional)(t.boolean),
    listenToFailureStates: (0, optional_1.Optional)(t.boolean),
    txRequestStates: (0, optional_1.Optional)(t.array(transactionRequestState_1.TransactionRequestState)),
    txRequestTransactionStates: (0, optional_1.Optional)(t.array(transactionState_1.TransactionState)),
    customHttpHeaders: (0, optional_1.Optional)(t.record(t.string, t.string)),
};
exports.AddWalletWebhookRequestBodyC = t.type({
    ...exports.AddWalletWebhookRequestBody,
});
