// @bitgo/public-types/dist/src/schema/webhook/webhookType.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationWebhookType = exports.OrganizationWebhookTypeEnum = exports.EnterpriseWebhookType = exports.EnterpriseWebhookTypeEnum = exports.WalletWebhookType = exports.WalletWebhookTypeEnum = exports.WebhookType = exports.WebhookTypeEnum = void 0;
const t = __importStar(require("io-ts"));
var WebhookTypeEnum;
(function (WebhookTypeEnum) {
    WebhookTypeEnum["txRequest"] = "txRequest";
    WebhookTypeEnum["txRequestTransaction"] = "txRequestTransaction";
    WebhookTypeEnum["transfer"] = "transfer";
    WebhookTypeEnum["transaction"] = "transaction";
    WebhookTypeEnum["transactionRemoved"] = "transactionRemoved";
    WebhookTypeEnum["transactionExpire"] = "transactionExpire";
    WebhookTypeEnum["pendingapproval"] = "pendingapproval";
    WebhookTypeEnum["block"] = "block";
    WebhookTypeEnum["admin"] = "admin";
    WebhookTypeEnum["address_confirmation"] = "address_confirmation";
    WebhookTypeEnum["lowFee"] = "lowFee";
    WebhookTypeEnum["circuitBreaker"] = "circuitBreaker";
    WebhookTypeEnum["lowFeeAddressBalance"] = "lowFeeAddressBalance";
    WebhookTypeEnum["wallet_confirmation"] = "wallet_confirmation";
    WebhookTypeEnum["bankAccount"] = "bankAccount";
    WebhookTypeEnum["userKycState"] = "userKycState";
    WebhookTypeEnum["enterpriseKycState"] = "enterpriseKycState";
    WebhookTypeEnum["identityStatus"] = "identityStatus";
    WebhookTypeEnum["accessToken"] = "accessToken";
    WebhookTypeEnum["accountCreated"] = "accountCreated";
    WebhookTypeEnum["contractSigned"] = "contractSigned";
    WebhookTypeEnum["fundsDeposited"] = "fundsDeposited";
    WebhookTypeEnum["kycResult"] = "kycResult";
    WebhookTypeEnum["policyBalanceChange"] = "policyBalanceChange";
    WebhookTypeEnum["reversal"] = "reversal";
    WebhookTypeEnum["transaction_finality_on_l1"] = "transaction_finality_on_l1";
    WebhookTypeEnum["stuckTx"] = "stuckTx";
    WebhookTypeEnum["tradeOrder"] = "tradeOrder";
})(WebhookTypeEnum || (exports.WebhookTypeEnum = WebhookTypeEnum = {}));
exports.WebhookType = t.keyof(WebhookTypeEnum, "WebhookType");
var WalletWebhookTypeEnum;
(function (WalletWebhookTypeEnum) {
    WalletWebhookTypeEnum["txRequest"] = "txRequest";
    WalletWebhookTypeEnum["txRequestTransaction"] = "txRequestTransaction";
    WalletWebhookTypeEnum["transfer"] = "transfer";
    WalletWebhookTypeEnum["transaction"] = "transaction";
    WalletWebhookTypeEnum["transactionRemoved"] = "transactionRemoved";
    WalletWebhookTypeEnum["transactionExpire"] = "transactionExpire";
    WalletWebhookTypeEnum["pendingapproval"] = "pendingapproval";
    WalletWebhookTypeEnum["block"] = "block";
    WalletWebhookTypeEnum["admin"] = "admin";
    WalletWebhookTypeEnum["address_confirmation"] = "address_confirmation";
    WalletWebhookTypeEnum["lowFee"] = "lowFee";
    WalletWebhookTypeEnum["circuitBreaker"] = "circuitBreaker";
    WalletWebhookTypeEnum["lowFeeAddressBalance"] = "lowFeeAddressBalance";
    WalletWebhookTypeEnum["transaction_finality_on_l1"] = "transaction_finality_on_l1";
    WalletWebhookTypeEnum["fundsDeposited"] = "fundsDeposited";
    WalletWebhookTypeEnum["stuckTx"] = "stuckTx";
})(WalletWebhookTypeEnum || (exports.WalletWebhookTypeEnum = WalletWebhookTypeEnum = {}));
exports.WalletWebhookType = t.keyof(WalletWebhookTypeEnum, "WalletWebhookType");
var EnterpriseWebhookTypeEnum;
(function (EnterpriseWebhookTypeEnum) {
    EnterpriseWebhookTypeEnum["bankAccount"] = "bankAccount";
    EnterpriseWebhookTypeEnum["accessToken"] = "accessToken";
    EnterpriseWebhookTypeEnum["policyBalanceChange"] = "policyBalanceChange";
    EnterpriseWebhookTypeEnum["reversal"] = "reversal";
})(EnterpriseWebhookTypeEnum || (exports.EnterpriseWebhookTypeEnum = EnterpriseWebhookTypeEnum = {}));
exports.EnterpriseWebhookType = t.keyof(EnterpriseWebhookTypeEnum, "EnterpriseWebhookType");
var OrganizationWebhookTypeEnum;
(function (OrganizationWebhookTypeEnum) {
    OrganizationWebhookTypeEnum["transfer"] = "transfer";
    OrganizationWebhookTypeEnum["txRequest"] = "txRequest";
    OrganizationWebhookTypeEnum["pendingapproval"] = "pendingapproval";
    OrganizationWebhookTypeEnum["txRequestTransaction"] = "txRequestTransaction";
    OrganizationWebhookTypeEnum["bankAccount"] = "bankAccount";
    OrganizationWebhookTypeEnum["identityStatus"] = "identityStatus";
    OrganizationWebhookTypeEnum["userKycState"] = "userKycState";
    OrganizationWebhookTypeEnum["enterpriseKycState"] = "enterpriseKycState";
    OrganizationWebhookTypeEnum["accessToken"] = "accessToken";
    OrganizationWebhookTypeEnum["reversal"] = "reversal";
    OrganizationWebhookTypeEnum["tradeOrder"] = "tradeOrder";
})(OrganizationWebhookTypeEnum || (exports.OrganizationWebhookTypeEnum = OrganizationWebhookTypeEnum = {}));
exports.OrganizationWebhookType = t.keyof(OrganizationWebhookTypeEnum, "OrganizationWebhookType");
