// @bitgo/public-types/dist/src/schema/webhook/webhook.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhook = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const transactionRequestState_1 = require("../transactionRequest/transactionRequestState");
const transactionState_1 = require("../transactionRequest/transactionState");
const webhookState_1 = require("./webhookState");
const webhookType_1 = require("./webhookType");
const webhookScope_1 = require("./webhookScope");
exports.Webhook = t.intersection([
    t.type({
        id: t.string,
        created: io_ts_types_1.DateFromISOString,
        coin: t.string,
        url: t.string,
        version: t.number,
        scope: webhookScope_1.WebhookScope,
        state: webhookState_1.WebhookState,
        successiveFailedAttempts: t.number,
        listenToFailureStates: t.boolean,
    }),
    t.partial({
        label: t.string,
        walletId: t.string,
        enterpriseId: t.string,
        organizationId: t.string,
        userId: t.string,
        type: webhookType_1.WebhookType,
        numConfirmations: t.number,
        lastAttempt: io_ts_types_1.DateFromISOString,
        failingSince: io_ts_types_1.DateFromISOString,
        allToken: t.boolean,
        txRequestStates: t.array(transactionRequestState_1.TransactionRequestState),
        txRequestTransactionStates: t.array(transactionState_1.TransactionState),
    }),
]);
