// @bitgo/sdk-coin-tao/dist/src/lib/moveStakeBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoveStakeBuilder = void 0;
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const constants_1 = require("./constants");
const moveStakeTransaction_1 = require("./moveStakeTransaction");
class MoveStakeBuilder extends abstract_substrate_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new moveStakeTransaction_1.MoveStakeTransaction(_coinConfig);
    }
    /**
     * Construct a transaction to move stake
     * @returns {UnsignedTransaction} an unsigned move stake transaction
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.moveStake({
            originHotkey: this._originHotkey,
            destinationHotkey: this._destinationHotkey,
            originNetuid: this._originNetuid,
            destinationNetuid: this._destinationNetuid,
            alphaAmount: this._alphaAmount,
        }, baseTxInfo);
    }
    /** @inheritdoc */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingRedelegate;
    }
    /**
     * Set the amount to move
     * @param {string} amount to move
     * @returns {MoveStakeBuilder} This builder.
     */
    amount(amount) {
        const value = new bignumber_js_1.default(amount);
        this.validateAmount(value);
        this._alphaAmount = amount;
        return this;
    }
    /**
     * Set the origin hot key address
     * @param {string} address of origin hotkey
     * @returns {MoveStakeBuilder} This builder.
     */
    originHotkey({ address }) {
        this.validateAddress({ address });
        this._originHotkey = address;
        return this;
    }
    /**
     * Set the destination hot key address
     * @param {string} address of destination hotkey
     * @returns {MoveStakeBuilder} This builder.
     */
    destinationHotkey({ address }) {
        this.validateAddress({ address });
        this._destinationHotkey = address;
        return this;
    }
    /**
     * Set the origin netuid of the subnet (root network is 0)
     * @param {string} netuid of subnet
     * @returns {MoveStakeBuilder} This builder.
     */
    originNetuid(netuid) {
        this.validateNetuid(netuid);
        this._originNetuid = netuid;
        return this;
    }
    /**
     * Set the destination netuid of the subnet (root network is 0)
     * @param {string} netuid of subnet
     * @returns {MoveStakeBuilder} This builder.
     */
    destinationNetuid(netuid) {
        this.validateNetuid(netuid);
        this._destinationNetuid = netuid;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name !== abstract_substrate_1.Interface.MethodNames.MoveStake) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected ${abstract_substrate_1.Interface.MethodNames.MoveStake}`);
        }
        const txMethod = this._method.args;
        this.amount(txMethod.alphaAmount);
        this.originHotkey({ address: txMethod.originHotkey });
        this.destinationHotkey({ address: txMethod.destinationHotkey });
        this.originNetuid(txMethod.originNetuid);
        this.destinationNetuid(txMethod.destinationNetuid);
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._originHotkey, this._destinationHotkey, this._originNetuid, this._destinationNetuid, this._alphaAmount);
    }
    /**
     * @param {BigNumber} amount amount to validate
     * @throws {InvalidTransactionError} if amount is less than or equal to zero
     */
    validateAmount(amount) {
        if (amount.isLessThanOrEqualTo(0)) {
            throw new sdk_core_1.InvalidTransactionError('Amount must be greater than zero');
        }
    }
    /**
     * @param {string} netuid netuid to validate
     * @throws {InvalidTransactionError} if netuid is out of range
     */
    validateNetuid(netuid) {
        const trimmed = netuid.trim();
        if (!/^\d+$/.test(trimmed)) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid netuid: ${netuid}. Must be a non-negative integer.`);
        }
        const num = Number(trimmed);
        if (num < 0 || num > constants_1.TAO_CONSTANTS.MAX_NETUID) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid netuid: ${netuid}. Netuid must be between 0 and ${constants_1.TAO_CONSTANTS.MAX_NETUID}.`);
        }
    }
    /**
     * Helper method to validate whether tx params have the correct type and format
     * @param {string} originHotkey origin hotkey address
     * @param {string} destinationHotkey destination hotkey address
     * @param {string} originNetuid netuid of the origin subnet
     * @param {string} destinationNetuid netuid of the destination subnet
     * @param {string} alphaAmount amount to move
     * @throws {InvalidTransactionError} if validation fails
     */
    validateFields(originHotkey, destinationHotkey, originNetuid, destinationNetuid, alphaAmount) {
        // Validate netuid ranges
        this.validateNetuid(originNetuid);
        this.validateNetuid(destinationNetuid);
        const validationResult = abstract_substrate_1.Schema.MoveStakeTransactionSchema.validate({
            originHotkey,
            destinationHotkey,
            originNetuid,
            destinationNetuid,
            alphaAmount,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Transaction validation failed: ${validationResult.error.message}`);
        }
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === abstract_substrate_1.Interface.MethodNames.MoveStake) {
            const txMethod = decodedTxn.method.args;
            const validationResult = abstract_substrate_1.Schema.MoveStakeTransactionSchema.validate(txMethod);
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Move Stake Transaction validation failed: ${validationResult.error.message}`);
            }
        }
    }
    /**
     * Construct a transaction to move stake
     *
     * @param {Interface.MoveStakeArgs} args arguments to be passed to the moveStake method
     * @param {Interface.CreateBaseTxInfo} info txn info required to construct the moveStake txn
     * @returns {UnsignedTransaction} an unsigned move stake transaction
     */
    moveStake(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'moveStake',
                pallet: 'subtensorModule',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.MoveStakeBuilder = MoveStakeBuilder;
