// @bitgo/sdk-coin-tao/dist/src/lib/tokenTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const tokenTransferTransaction_1 = require("./tokenTransferTransaction");
class TokenTransferBuilder extends abstract_substrate_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new tokenTransferTransaction_1.TokenTransferTransaction(_coinConfig);
    }
    /**
     * Construct a transaction to transfer stake
     * @returns {UnsignedTransaction} an unsigned transfer stake transaction
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.transferStake({
            destinationColdkey: this._destinationColdkey,
            hotkey: this._hotkey,
            originNetuid: this._originNetuid,
            destinationNetuid: this._destinationNetuid,
            alphaAmount: this._alphaAmount,
        }, baseTxInfo);
    }
    /** @inheritdoc */
    get transactionType() {
        return sdk_core_1.TransactionType.SendToken;
    }
    /**
     * Set the amount to transfer
     * @param {string} amount to transfer
     * @returns {TokenTransferBuilder} This builder.
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._alphaAmount = amount;
        return this;
    }
    /**
     * Set the validator hot key address
     * @param {string} hotkey address of validator
     * @returns {TokenTransferBuilder} This builder.
     */
    hotkey(address) {
        this.validateAddress({ address });
        this._hotkey = address;
        return this;
    }
    /**
     * Set the destination cold key address
     * @param {string} address of the destination cold key
     * @returns {TokenTransferBuilder} This builder.
     */
    destinationColdkey(address) {
        this.validateAddress({ address });
        this._destinationColdkey = address;
        return this;
    }
    /**
     * Set the origin netuid of the subnet (root network is 0)
     * @param {string} netuid of subnet
     * @returns {TokenTransferBuilder} This builder.
     */
    originNetuid(netuid) {
        this.validateNetuid(netuid);
        this._originNetuid = netuid;
        return this;
    }
    /**
     * Set the destination netuid of the subnet (root network is 0)
     * @param {string} netuid of subnet
     * @returns {TokenTransferBuilder} This builder.
     */
    destinationNetuid(netuid) {
        this.validateNetuid(netuid);
        this._destinationNetuid = netuid;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name === abstract_substrate_1.Interface.MethodNames.TransferStake) {
            const txMethod = this._method.args;
            this.amount(txMethod.alphaAmount);
            this.hotkey(txMethod.hotkey);
            this.destinationColdkey(txMethod.destinationColdkey);
            this.originNetuid(txMethod.originNetuid);
            this.destinationNetuid(txMethod.destinationNetuid);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected ${abstract_substrate_1.Interface.MethodNames.TransferStake}`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._destinationColdkey, this._hotkey, this._originNetuid, this._destinationNetuid, this._alphaAmount);
    }
    /**
     * Helper method to validate whether tx params have the correct type and format
     * @param {string} destinationColdkey destination cold key address
     * @param {string} hotkey hotkey address of the validator
     * @param {string} originNetuid netuid of the origin subnet
     * @param {string} destinationNetuid netuid of the destination subnet
     * @param {string} alphaAmount amount to transfer
     * @throws {InvalidTransactionError} if validation fails
     */
    validateFields(destinationColdkey, hotkey, originNetuid, destinationNetuid, alphaAmount) {
        const validationResult = abstract_substrate_1.Schema.TransferStakeTransactionSchema.validate({
            destinationColdkey,
            hotkey,
            originNetuid,
            destinationNetuid,
            alphaAmount,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Transaction validation failed: ${validationResult.error.message}`);
        }
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === abstract_substrate_1.Interface.MethodNames.TransferStake) {
            const txMethod = decodedTxn.method.args;
            const validationResult = abstract_substrate_1.Schema.TransferStakeTransactionSchema.validate(txMethod);
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Transfer Transaction validation failed: ${validationResult.error.message}`);
            }
        }
    }
    /**
     * Validate the netuid. Throws an error if the netuid is not a non-negative integer.
     * @param netuid The netuid to validate.
     */
    validateNetuid(netuid) {
        if (BigInt(netuid) < 0n) {
            throw new Error(`The netuid '${netuid}' is not a valid netuid`);
        }
    }
    /**
     * Construct a transaction to transfer stake
     *
     * @param {Interface.TransferStakeArgs} args arguments to be passed to the transferStake method
     * @param {Interface.CreateBaseTxInfo} info txn info required to construct the transferStake txn
     * @returns {UnsignedTransaction} an unsigned stake transaction
     */
    transferStake(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'transferStake',
                pallet: 'subtensorModule',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
