// @bitgo/sdk-coin-tao/dist/src/lib/tokenTransferTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferTransaction = void 0;
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const utils_1 = __importDefault(require("./utils"));
class TokenTransferTransaction extends abstract_substrate_1.Transaction {
    /** @inheritdoc */
    toJson() {
        if (!this._substrateTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        const decodedTx = (0, txwrapper_polkadot_1.decode)(this._substrateTransaction, {
            metadataRpc: this._substrateTransaction.metadataRpc,
            registry: this._registry,
            isImmortalEra: utils_1.default.isZeroHex(this._substrateTransaction.era),
        });
        const txMethod = decodedTx.method.args;
        const result = super.toJson();
        result.destinationColdkey = txMethod.destinationColdkey;
        result.hotkey = txMethod.hotkey;
        result.originNetuid = txMethod.originNetuid;
        result.destinationNetuid = txMethod.destinationNetuid;
        result.alphaAmount = txMethod.alphaAmount;
        return result;
    }
    /** @inheritdoc */
    explainTransaction() {
        const result = this.toJson();
        const outputs = this._outputs.map((output) => {
            return {
                address: output.address,
                amount: output.value,
                tokenName: output.coin,
            };
        });
        const explanationResult = {
            id: result.id,
            outputAmount: result.amount?.toString() || '0',
            changeAmount: '0',
            changeOutputs: [],
            outputs,
            fee: {
                fee: result.tip?.toString() || '',
                type: 'tip',
            },
            type: this.type,
        };
        return explanationResult;
    }
}
exports.TokenTransferTransaction = TokenTransferTransaction;
