// @bitgo/sdk-coin-tao/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const _1 = require("./");
const transferBuilder_1 = require("./transferBuilder");
const utils_1 = __importDefault(require("./utils"));
const stakingBuilder_1 = require("./stakingBuilder");
const unstakeBuilder_1 = require("./unstakeBuilder");
const tokenTransferBuilder_1 = require("./tokenTransferBuilder");
const moveStakeBuilder_1 = require("./moveStakeBuilder");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._material = utils_1.default.getMaterial(_coinConfig.network.type);
    }
    getTransferBuilder() {
        return new transferBuilder_1.TransferBuilder(this._coinConfig).material(this._material);
    }
    getStakingBuilder() {
        return new stakingBuilder_1.StakingBuilder(this._coinConfig).material(this._material);
    }
    getUnstakingBuilder() {
        return new unstakeBuilder_1.UnstakeBuilder(this._coinConfig).material(this._material);
    }
    getTokenTransferBuilder() {
        return new tokenTransferBuilder_1.TokenTransferBuilder(this._coinConfig).material(this._material);
    }
    getMoveStakeBuilder() {
        return new moveStakeBuilder_1.MoveStakeBuilder(this._coinConfig).material(this._material);
    }
    getWalletInitializationBuilder() {
        throw new sdk_core_1.NotImplementedError(`walletInitialization for ${this._coinConfig.name} not implemented`);
    }
    from(rawTxn) {
        const builder = this.getBuilder(rawTxn);
        builder.from(rawTxn);
        return builder;
    }
    material(material) {
        this._material = material;
        return this;
    }
    getBuilder(rawTxn) {
        const registry = _1.SingletonRegistry.getInstance(this._material);
        const decodedTxn = (0, txwrapper_polkadot_1.decode)(rawTxn, {
            metadataRpc: this._material.metadata,
            registry: registry,
        });
        const methodName = decodedTxn.method?.name;
        if (methodName === _1.Interface.MethodNames.TransferKeepAlive || methodName === _1.Interface.MethodNames.TransferAll) {
            return this.getTransferBuilder();
        }
        else if (methodName === _1.Interface.MethodNames.AddStake) {
            return this.getStakingBuilder();
        }
        else if (methodName === _1.Interface.MethodNames.RemoveStake) {
            return this.getUnstakingBuilder();
        }
        else if (methodName === _1.Interface.MethodNames.TransferStake) {
            return this.getTokenTransferBuilder();
        }
        else if (methodName === _1.Interface.MethodNames.MoveStake) {
            return this.getMoveStakeBuilder();
        }
        else {
            throw new sdk_core_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
