// @bitgo/sdk-coin-tao/dist/src/lib/unstakeBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnstakeBuilder = void 0;
const txwrapper_core_1 = require("@substrate/txwrapper-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
class UnstakeBuilder extends abstract_substrate_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * Construct a transaction to unstake
     *
     * @returns {UnsignedTransaction} an unsigned unstake TAO transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#staking
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.removeStake({
            amountUnstaked: this._amount,
            hotkey: this._hotkey,
            netuid: this._netuid,
        }, baseTxInfo);
    }
    /** @inheritdoc */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingDeactivate;
    }
    /**
     * The amount to unstake.
     *
     * @param {string} amount to unstake
     * @returns {UnstakeBuilder} This unstaking builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-nominator#required-minimum-stake
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     * The controller of the staked amount.
     *
     * @param {string} hotkey address of validator
     * @returns {UnstakeBuilder} This unstaking builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-staking#accounts
     */
    hotkey({ address }) {
        this.validateAddress({ address });
        this._hotkey = address;
        return this;
    }
    /**
     * Netuid of the subnet (root network is 0)
     * @param {string} netuid
     * @returns {UnstakeBuilder} This unstaking builder
     */
    netuid(netuid) {
        this._netuid = netuid;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name === abstract_substrate_1.Interface.MethodNames.RemoveStake) {
            const txMethod = this._method.args;
            this.amount(txMethod.amountUnstaked);
            this.hotkey({ address: txMethod.hotkey });
            this.netuid(txMethod.netuid);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected addStake`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._amount, this._hotkey, this._netuid);
    }
    /**
     * Helper method to validate whether unstake params have the correct type and format
     * @param {string} amountUnstaked amount to unstake
     * @param {string} hotkey hotkey address of the validator
     * @param {string} netuid netuid of the subnet
     */
    validateFields(amountUnstaked, hotkey, netuid) {
        const validationResult = abstract_substrate_1.Schema.UnstakeTransactionSchema.validate({
            amountUnstaked,
            hotkey,
            netuid,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`UnStake Builder Transaction validation failed: ${validationResult.error.message}`);
        }
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === abstract_substrate_1.Interface.MethodNames.RemoveStake) {
            const txMethod = decodedTxn.method.args;
            const amountUnstaked = txMethod.amountUnstaked;
            const hotkey = txMethod.hotkey;
            const netuid = txMethod.netuid;
            const validationResult = abstract_substrate_1.Schema.UnstakeTransactionSchema.validate({ amountUnstaked, hotkey, netuid });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Transfer Transaction validation failed: ${validationResult.error.message}`);
            }
        }
    }
    /**
     * Construct a transaction to unstake
     *
     * @param {Interface.RemoveStakeArgs} RemoveStake arguments to be passed to the addStake method
     * @param {Interface.CreateBaseTxInfo} Base txn info required to construct the removeStake txn
     * @returns {UnsignedTransaction} an unsigned unstake TAO transaction
     */
    removeStake(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'removeStake',
                pallet: 'subtensorModule',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.UnstakeBuilder = UnstakeBuilder;
