// @bitgo/sdk-core/dist/test/unit/bitgo/wallet/BuildParams.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const BuildParams_1 = require("../../../../src/bitgo/wallet/BuildParams");
describe('BuildParams', function () {
    it('enforces codec', function () {
        assert.deepStrictEqual(BuildParams_1.BuildParams.encode({
            addressType: 'p2sh',
            txFormat: 'psbt',
            isReplaceableByFee: true,
            recipients: [
                {
                    amount: '10000',
                    address: '2N9Ego9KidiZR8tMP82g6RaggQtcbR9zNzH',
                },
            ],
        }), {
            addressType: 'p2sh',
            txFormat: 'psbt',
            isReplaceableByFee: true,
            recipients: [
                {
                    amount: '10000',
                    address: '2N9Ego9KidiZR8tMP82g6RaggQtcbR9zNzH',
                },
            ],
        });
        assert.deepStrictEqual(BuildParams_1.BuildParams.encode({
            rbfTxIds: ['tx1'],
        }), {
            rbfTxIds: ['tx1'],
        });
    });
});
