// @bitgo/sdk-core/dist/test/unit/bitgo/wallet/getAccountResources.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const sinon = __importStar(require("sinon"));
require("should");
const src_1 = require("../../../../src");
describe('Wallet - getAccountResources', function () {
    let wallet;
    let mockBitGo;
    let mockBaseCoin;
    let mockWalletData;
    beforeEach(function () {
        mockBitGo = {
            get: sinon.stub(),
        };
        mockBaseCoin = {
            url: sinon.stub().returns('/test/coin'),
            supportsTss: sinon.stub().returns(false),
        };
        mockWalletData = {
            id: 'test-wallet-id',
            keys: ['user-key', 'backup-key', 'bitgo-key'],
        };
        wallet = new src_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
    });
    afterEach(function () {
        sinon.restore();
    });
    describe('getAccountResources', function () {
        it('should call WP API with addresses parameter', async function () {
            const mockResponse = {
                resources: [
                    { address: 'address1', balance: 100 },
                    { address: 'address2', balance: 200 },
                ],
            };
            mockBitGo.get.returns({
                query: sinon.stub().returns({
                    result: sinon.stub().resolves(mockResponse),
                }),
            });
            const addresses = ['address1', 'address2'];
            const result = await wallet.getAccountResources({ addresses });
            result.should.deepEqual(mockResponse);
            sinon.assert.calledOnce(mockBitGo.get);
            const queryStub = mockBitGo.get.returnValues[0].query;
            sinon.assert.calledWith(queryStub, { addresses });
        });
        it('should call WP API with addresses and destinationAddress parameters', async function () {
            const mockResponse = {
                resources: [{ address: 'address1', balance: 100 }],
            };
            mockBitGo.get.returns({
                query: sinon.stub().returns({
                    result: sinon.stub().resolves(mockResponse),
                }),
            });
            const addresses = ['address1'];
            const destinationAddress = 'TDestAddress123';
            const result = await wallet.getAccountResources({ addresses, destinationAddress });
            result.should.deepEqual(mockResponse);
            sinon.assert.calledOnce(mockBitGo.get);
            const queryStub = mockBitGo.get.returnValues[0].query;
            sinon.assert.calledWith(queryStub, { addresses, destinationAddress });
        });
        it('should throw error if addresses is not an array', async function () {
            try {
                await wallet.getAccountResources({ addresses: 'not-an-array' });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('addresses must be an array');
            }
        });
        it('should throw error if addresses array is empty', async function () {
            try {
                await wallet.getAccountResources({ addresses: [] });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('addresses array cannot be empty');
            }
        });
        it('should not include destinationAddress in query if not provided', async function () {
            const mockResponse = { resources: [] };
            mockBitGo.get.returns({
                query: sinon.stub().returns({
                    result: sinon.stub().resolves(mockResponse),
                }),
            });
            const addresses = ['address1'];
            await wallet.getAccountResources({ addresses });
            const queryStub = mockBitGo.get.returnValues[0].query;
            const queryArg = queryStub.firstCall.args[0];
            queryArg.should.deepEqual({ addresses });
            queryArg.should.not.have.property('destinationAddress');
        });
    });
});
