// @bitgo/sdk-core/dist/test/unit/bitgo/wallet/walletEvmAddressCreation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const sinon = __importStar(require("sinon"));
require("should");
const wallet_1 = require("../../../../src/bitgo/wallet/wallet");
describe('Wallet - EVM Keyring Address Creation', function () {
    let wallet;
    let mockBitGo;
    let mockBaseCoin;
    let mockWalletData;
    beforeEach(function () {
        mockBitGo = {
            post: sinon.stub(),
            setRequestTracer: sinon.stub(),
        };
        mockBaseCoin = {
            isEVM: sinon.stub(),
            supportsTss: sinon.stub().returns(true),
            getFamily: sinon.stub().returns('eth'),
            isValidAddress: sinon.stub(),
            keychains: sinon.stub(),
            url: sinon.stub().returns('/test/wallet/address'),
        };
        mockWalletData = {
            id: 'test-wallet-id',
            keys: ['user-key', 'backup-key', 'bitgo-key'],
        };
        wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
    });
    afterEach(function () {
        sinon.restore();
    });
    describe('createAddress with EVM keyring parameters', function () {
        beforeEach(function () {
            mockBaseCoin.isEVM.returns(true);
            mockBaseCoin.isValidAddress.returns(true);
            mockBaseCoin.keychains.returns({
                get: sinon.stub().resolves({ id: 'keychain-id', pub: 'public-key' }),
            });
        });
        it('should create address with evmKeyRingReferenceAddress', async function () {
            const mockAddressResponse = {
                id: '507f1f77bcf86cd799439012',
                address: '0x1234567890123456789012345678901234567890',
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            const result = await wallet.createAddress({
                chain: 0,
                label: 'Test EVM Address',
                evmKeyRingReferenceAddress: '0x742d35Cc6634C0532925a3b8D404fddF4f780EAD',
            });
            result.should.have.property('id', '507f1f77bcf86cd799439012');
            result.should.have.property('address', '0x1234567890123456789012345678901234567890');
            mockBitGo.post.should.have.been.calledOnce;
        });
        it('should throw error if evmKeyRingReferenceAddress is not a string', async function () {
            try {
                await wallet.createAddress({
                    chain: 0,
                    label: 'Test Address',
                    evmKeyRingReferenceAddress: 123,
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceAddress has to be a string');
            }
        });
        it('should throw error if evmKeyRingReferenceAddress is not a valid address', async function () {
            mockBaseCoin.isValidAddress.returns(false);
            try {
                await wallet.createAddress({
                    chain: 0,
                    label: 'Test Address',
                    evmKeyRingReferenceAddress: 'invalid-address',
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceAddress must be a valid address');
            }
        });
        it('should throw error for non-EVM chains with evmKeyRingReferenceAddress', async function () {
            mockBaseCoin.isEVM.returns(false);
            try {
                await wallet.createAddress({
                    chain: 0,
                    label: 'Test Address',
                    evmKeyRingReferenceAddress: '1BvBMSEYstWetqTFn5Au4m4GFg7xJaNVN2',
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceAddress is only supported for EVM chains');
            }
        });
        it('should create address without reference parameters for regular addresses', async function () {
            const mockAddressResponse = {
                id: 'regular-address-id',
                address: '0x9876543210987654321098765432109876543210',
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            const result = await wallet.createAddress({
                chain: 0,
                label: 'Regular Address',
            });
            result.should.have.property('id', 'regular-address-id');
            result.should.have.property('address', '0x9876543210987654321098765432109876543210');
            mockBitGo.post.should.have.been.calledOnce;
        });
    });
    describe('Non-EVM chains', function () {
        beforeEach(function () {
            mockBaseCoin.isEVM.returns(false);
        });
        it('should create regular addresses for non-EVM chains', async function () {
            const mockAddressResponse = {
                id: 'btc-address-id',
                address: '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa',
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.keychains.returns({
                get: sinon.stub().resolves({ id: 'keychain-id', pub: 'public-key' }),
            });
            const result = await wallet.createAddress({
                chain: 0,
                label: 'BTC Address',
            });
            result.should.have.property('id', 'btc-address-id');
            result.should.have.property('address', '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa');
            mockBitGo.post.should.have.been.calledOnce;
        });
    });
});
