// @bitgo/sdk-core/dist/test/unit/bitgo/wallet/walletsEvmKeyring.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const sinon = __importStar(require("sinon"));
require("should");
const wallets_1 = require("../../../../src/bitgo/wallet/wallets");
describe('Wallets', function () {
    let wallets;
    let mockBitGo;
    let mockBaseCoin;
    beforeEach(function () {
        mockBitGo = {
            post: sinon.stub(),
            encrypt: sinon.stub(),
            setRequestTracer: sinon.stub(),
        };
        mockBaseCoin = {
            isEVM: sinon.stub(),
            supportsTss: sinon.stub().returns(true),
            getFamily: sinon.stub().returns('eth'),
            getDefaultMultisigType: sinon.stub(),
            keychains: sinon.stub(),
            url: sinon.stub().returns('/test/url'),
            isValidMofNSetup: sinon.stub(),
            getConfig: sinon.stub().returns({ features: [] }),
        };
        wallets = new wallets_1.Wallets(mockBitGo, mockBaseCoin);
    });
    afterEach(function () {
        sinon.restore();
    });
    describe('EVM Keyring - generateWallet', function () {
        beforeEach(function () {
            mockBaseCoin.isEVM.returns(true);
            mockBaseCoin.supportsTss.returns(true);
            mockBaseCoin.getDefaultMultisigType.returns('tss');
        });
        it('should create EVM wallet with evmKeyRingReferenceWalletId', async function () {
            const mockWalletResponse = {
                id: '597f1f77bcf86cd799439011',
                keys: ['user-key', 'backup-key', 'bitgo-key'],
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockWalletResponse),
                }),
            });
            mockBaseCoin.keychains.returns({
                get: sinon.stub().resolves({ id: 'keychain-id', pub: 'public-key' }),
            });
            const result = await wallets.generateWallet({
                label: 'Test EVM Wallet',
                evmKeyRingReferenceWalletId: '507f1f77bcf86cd799439011',
            });
            result.should.have.property('wallet');
            result.should.have.property('userKeychain');
            result.should.have.property('backupKeychain');
            result.should.have.property('bitgoKeychain');
        });
        it('should throw error if evmKeyRingReferenceWalletId provided for non-EVM chain', async function () {
            mockBaseCoin.isEVM.returns(false);
            try {
                await wallets.generateWallet({
                    label: 'Test Wallet',
                    evmKeyRingReferenceWalletId: '507f1f77bcf86cd799439011',
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceWalletId is only supported for EVM chains');
            }
        });
        it('should throw error if evmKeyRingReferenceWalletId is not a string', async function () {
            try {
                await wallets.generateWallet({
                    label: 'Test Wallet',
                    evmKeyRingReferenceWalletId: 123,
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('invalid evmKeyRingReferenceWalletId argument, expecting string');
            }
        });
    });
    describe('EVM Keyring - add method', function () {
        beforeEach(function () {
            mockBaseCoin.isEVM.returns(true);
        });
        it('should add EVM wallet with evmKeyRingReferenceWalletId without multisig validation', async function () {
            const mockWalletResponse = {
                id: 'new-wallet-id',
                keys: ['user-key', 'backup-key', 'bitgo-key'],
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockWalletResponse),
                }),
            });
            const result = await wallets.add({
                label: 'Test EVM Wallet',
                evmKeyRingReferenceWalletId: 'reference-wallet-id',
            });
            result.should.have.property('wallet');
            mockBitGo.post.should.have.been.calledOnce;
        });
        it('should skip multisig validation for EVM wallets with evmKeyRingReferenceWalletId', async function () {
            const mockWalletResponse = {
                id: 'new-wallet-id',
                keys: ['user-key', 'backup-key', 'bitgo-key'],
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockWalletResponse),
                }),
            });
            // This should not throw error even without keys, m, n parameters
            const result = await wallets.add({
                label: 'Test EVM Wallet',
                evmKeyRingReferenceWalletId: '507f1f77bcf86cd799439011',
                // No keys, m, n provided - should be fine for EVM keyring
            });
            result.should.have.property('wallet');
        });
        it('should throw error for non-EVM chains with evmKeyRingReferenceWalletId', async function () {
            mockBaseCoin.isEVM.returns(false);
            try {
                await wallets.add({
                    label: 'Test Wallet',
                    evmKeyRingReferenceWalletId: '507f1f77bcf86cd799439011',
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceWalletId is only supported for EVM chains');
            }
        });
        it('should still validate multisig for regular wallets', async function () {
            mockBaseCoin.isEVM.returns(true);
            try {
                await wallets.add({
                    label: 'Test Wallet',
                    type: 'hot',
                    // No evmKeyRingReferenceWalletId, so should require keys, m, n
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('invalid argument');
            }
        });
    });
    describe('Non-EVM chains', function () {
        beforeEach(function () {
            mockBaseCoin.isEVM.returns(false);
        });
        it('should not allow evmKeyRingReferenceWalletId for non-EVM chains in generateWallet', async function () {
            try {
                await wallets.generateWallet({
                    label: 'Test Wallet',
                    evmKeyRingReferenceWalletId: 'reference-wallet-id',
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceWalletId is only supported for EVM chains');
            }
        });
        it('should not allow evmKeyRingReferenceWalletId for non-EVM chains in add', async function () {
            try {
                await wallets.add({
                    label: 'Test Wallet',
                    evmKeyRingReferenceWalletId: 'reference-wallet-id',
                });
                assert.fail('Should have thrown error');
            }
            catch (error) {
                error.message.should.equal('evmKeyRingReferenceWalletId is only supported for EVM chains');
            }
        });
    });
});
