// @bitgo/sdk-core/dist/test/unit/bitgo/wallet/walletTssAddressVerification.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const sinon = __importStar(require("sinon"));
require("should");
const wallet_1 = require("../../../../src/bitgo/wallet/wallet");
const keychain_1 = require("../../../../src/bitgo/keychain");
describe('Wallet - TSS Address Verification with Derivation Prefix', function () {
    let wallet;
    let mockBitGo;
    let mockBaseCoin;
    let mockWalletData;
    let mockKeychains;
    beforeEach(function () {
        mockBitGo = {
            post: sinon.stub(),
            setRequestTracer: sinon.stub(),
        };
        mockBaseCoin = {
            supportsTss: sinon.stub().returns(true),
            getFamily: sinon.stub().returns('sol'),
            getMPCAlgorithm: sinon.stub().returns('eddsa'),
            isValidAddress: sinon.stub().returns(true),
            isWalletAddress: sinon.stub(),
            keychains: sinon.stub(),
            url: sinon.stub().returns('/api/v2/sol/wallet/test-wallet-id/address'),
        };
        // Default keychains setup - all with commonKeychain
        mockKeychains = [
            {
                id: 'user-key-id',
                type: 'tss',
                commonKeychain: '8ea32ecacfc83effbd2e2790ee44fa7c59b4d86c29a12f09fb613d8195f93f4e21875cad3b98adada40c040c54c3569467df41a020881a6184096378701862bd',
                derivedFromParentWithSeed: 'test-seed-user',
            },
            {
                id: 'backup-key-id',
                type: 'tss',
                commonKeychain: '8ea32ecacfc83effbd2e2790ee44fa7c59b4d86c29a12f09fb613d8195f93f4e21875cad3b98adada40c040c54c3569467df41a020881a6184096378701862bd',
            },
            {
                id: 'bitgo-key-id',
                type: 'tss',
                commonKeychain: '8ea32ecacfc83effbd2e2790ee44fa7c59b4d86c29a12f09fb613d8195f93f4e21875cad3b98adada40c040c54c3569467df41a020881a6184096378701862bd',
            },
        ];
        mockBaseCoin.keychains.returns({
            get: sinon.stub().callsFake((params) => {
                const keyIndex = mockWalletData.keys.indexOf(params.id);
                return Promise.resolve(mockKeychains[keyIndex]);
            }),
        });
        mockWalletData = {
            id: 'test-wallet-id',
            keys: ['user-key-id', 'backup-key-id', 'bitgo-key-id'],
            multisigType: 'tss',
            coinSpecific: {
                walletVersion: 3,
            },
        };
        wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
    });
    afterEach(function () {
        sinon.restore();
    });
    describe('Custodial TSS Wallet - No Derivation Prefix', function () {
        beforeEach(function () {
            mockWalletData.type = 'custodial';
            wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
        });
        it('should not extract derivation prefix for custodial wallets', async function () {
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {},
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.isWalletAddress.resolves(true);
            await wallet.createAddress({ chain: 0 });
            // Verify that custodial wallets don't use derivedFromParentWithSeed
            // (their commonKeychain already accounts for the prefix)
            const verificationCall = mockBaseCoin.isWalletAddress.getCall(0);
            const verificationData = verificationCall.args[0];
            assert.strictEqual(verificationData.derivedFromParentWithSeed, undefined);
        });
    });
    describe('Cold/SMC TSS Wallet - Derivation Prefix from User Keychain', function () {
        beforeEach(function () {
            mockWalletData.type = 'cold';
            wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
        });
        it('should extract derivation prefix from User keychain for cold/SMC wallets', async function () {
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {},
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.isWalletAddress.resolves(true);
            await wallet.createAddress({ chain: 0 });
            // Verify that isWalletAddress was called with derivedFromParentWithSeed from User keychain
            const verificationCall = mockBaseCoin.isWalletAddress.getCall(0);
            const verificationData = verificationCall.args[0];
            verificationData.should.have.property('derivedFromParentWithSeed');
            verificationData.derivedFromParentWithSeed.should.equal('test-seed-user');
        });
        it('should handle missing derivedFromParentWithSeed gracefully for cold wallet', async function () {
            // Remove derivedFromParentWithSeed from User keychain
            mockKeychains[keychain_1.KeyIndices.USER].derivedFromParentWithSeed = undefined;
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {},
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.isWalletAddress.resolves(true);
            await wallet.createAddress({ chain: 0 });
            // Should not have derivedFromParentWithSeed if seed is missing
            const verificationCall = mockBaseCoin.isWalletAddress.getCall(0);
            const verificationData = verificationCall.args[0];
            assert.strictEqual(verificationData.derivedFromParentWithSeed, undefined);
        });
    });
    describe('Non-TSS Wallet - No Derivation Prefix', function () {
        beforeEach(function () {
            mockWalletData.multisigType = 'onchain';
            mockWalletData.type = 'cold';
            wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
        });
        it('should not extract derivation prefix for non-TSS wallets', async function () {
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {},
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.isWalletAddress.resolves(true);
            await wallet.createAddress({ chain: 0 });
            // Verify that derivedFromParentWithSeed is not set for non-TSS wallets
            const verificationCall = mockBaseCoin.isWalletAddress.getCall(0);
            const verificationData = verificationCall.args[0];
            assert.strictEqual(verificationData.derivedFromParentWithSeed, undefined);
        });
    });
    describe('Edge Cases', function () {
        it('should handle wallet without USER keychain', async function () {
            // Set keys array to only have backup keychain (no USER keychain at index 0)
            mockWalletData.keys = ['backup-key-id']; // Only backup keychain, no USER
            mockWalletData.type = 'cold';
            // Update mock to return backup keychain when requested
            mockBaseCoin.keychains.returns({
                get: sinon.stub().callsFake((params) => {
                    if (params.id === 'backup-key-id') {
                        return Promise.resolve(mockKeychains[keychain_1.KeyIndices.BACKUP]);
                    }
                    return Promise.resolve(null);
                }),
            });
            wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {},
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.isWalletAddress.resolves(true);
            await wallet.createAddress({ chain: 0 });
            const verificationCall = mockBaseCoin.isWalletAddress.getCall(0);
            const verificationData = verificationCall.args[0];
            // Should not have derivedFromParentWithSeed if USER keychain doesn't exist in keys array
            assert.strictEqual(verificationData.derivedFromParentWithSeed, undefined);
        });
        it('should handle pendingChainInitialization correctly', async function () {
            mockWalletData.type = 'cold';
            wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {
                    pendingChainInitialization: true,
                },
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            // Should skip verification when pendingChainInitialization is true
            await wallet.createAddress({ chain: 0 });
            // isWalletAddress should not be called when pendingChainInitialization is true
            mockBaseCoin.isWalletAddress.should.not.have.been.called;
        });
        it('should handle forwarderVersion 1 with pendingChainInitialization', async function () {
            mockWalletData.type = 'cold';
            wallet = new wallet_1.Wallet(mockBitGo, mockBaseCoin, mockWalletData);
            const mockAddressResponse = {
                id: 'address-id',
                address: '6FjshVqwmDH74wfxkZrJaRGEjTeJQL4ViL6X18VXUNAY',
                index: 0,
                coinSpecific: {
                    pendingChainInitialization: true,
                    forwarderVersion: 1,
                },
            };
            mockBitGo.post.returns({
                send: sinon.stub().returns({
                    result: sinon.stub().resolves(mockAddressResponse),
                }),
            });
            mockBaseCoin.isWalletAddress.resolves(true);
            // Should verify even with pendingChainInitialization when forwarderVersion is 1
            await wallet.createAddress({ chain: 0, forwarderVersion: 1 });
            mockBaseCoin.isWalletAddress.should.have.been.calledOnce;
        });
    });
});
