// @bitgo/utxo-lib/dist/src/testutil/keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getKey = getKey;
exports.getKeyTriple = getKeyTriple;
exports.getUncompressedKeyTriple = getUncompressedKeyTriple;
exports.getKeyName = getKeyName;
exports.getDefaultCosigner = getDefaultCosigner;
exports.getWalletKeysForSeed = getWalletKeysForSeed;
exports.getDefaultWalletKeys = getDefaultWalletKeys;
const crypto = require("crypto");
const WalletKeys_1 = require("../bitgo/wallet/WalletKeys");
const secp256k1_1 = require("@bitgo/secp256k1");
const networks_1 = require("../networks");
const bip32 = (0, secp256k1_1.BIP32Factory)(secp256k1_1.ecc);
function getKey(seed) {
    return bip32.fromSeed(crypto.createHash('sha256').update(seed).digest());
}
function getKeyTriple(seed) {
    return [getKey(seed + '.0'), getKey(seed + '.1'), getKey(seed + '.2')];
}
function getUncompressedKey(input) {
    // Using input for deterministic randomness
    return secp256k1_1.ECPair.makeRandom({
        compressed: false,
        network: networks_1.networks.testnet,
        rng: () => {
            return Buffer.alloc(32, input);
        },
    });
}
function getUncompressedKeyTriple(inputs) {
    return [getUncompressedKey(inputs[0]), getUncompressedKey(inputs[1]), getUncompressedKey(inputs[2])];
}
function getKeyName(triple, k) {
    return ['user', 'backup', 'bitgo'][triple.indexOf(k)];
}
function getDefaultCosigner(keyset, signer) {
    const eq = (a, b) => a === b || (Buffer.isBuffer(a) && Buffer.isBuffer(b) && a.equals(b));
    const [user, backup, bitgo] = keyset;
    if (eq(signer, user)) {
        return bitgo;
    }
    if (eq(signer, backup)) {
        return bitgo;
    }
    if (eq(signer, bitgo)) {
        return user;
    }
    throw new Error(`signer not in pubkeys`);
}
function getWalletKeysForSeed(seed) {
    return new WalletKeys_1.RootWalletKeys(getKeyTriple(seed));
}
function getDefaultWalletKeys() {
    return getWalletKeysForSeed('default');
}
