// @bitgo/utxo-lib/dist/src/testutil/psbt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcidTest = exports.txFormats = exports.outputScriptTypes = exports.inputScriptTypes = exports.signStages = void 0;
exports.toUnspent = toUnspent;
exports.getSigners = getSigners;
exports.signPsbtInput = signPsbtInput;
exports.signAllPsbtInputs = signAllPsbtInputs;
exports.constructPsbt = constructPsbt;
exports.verifyFullySignedSignatures = verifyFullySignedSignatures;
const bitcoinjslib = require("bitcoinjs-lib");
const assert_1 = require("assert");
const outputScripts_1 = require("../bitgo/outputScripts");
const bitgo_1 = require("../bitgo");
const networks_1 = require("../networks");
const mock_1 = require("./mock");
const address_1 = require("../address");
const keys_1 = require("./keys");
exports.signStages = ['unsigned', 'halfsigned', 'fullsigned'];
/**
 * array of supported input script types.
 * use p2trMusig2 for p2trMusig2 script path.
 * use taprootKeyPathSpend for p2trMusig2 key path.
 */
exports.inputScriptTypes = [...outputScripts_1.scriptTypes2Of3, 'taprootKeyPathSpend', outputScripts_1.scriptTypeP2shP2pk];
/**
 * array of supported output script types.
 */
exports.outputScriptTypes = outputScripts_1.scriptTypes2Of3;
/**
 * create unspent object from input script type, index, network and root wallet key.
 */
function toUnspent(input, index, network, rootWalletKeys, { p2shP2pkKey } = {}) {
    if (input.scriptType === 'p2shP2pk') {
        return (0, mock_1.mockReplayProtectionUnspent)(network, input.value, {
            key: p2shP2pkKey ?? rootWalletKeys['user'],
            vout: index,
        });
    }
    else {
        const chain = (0, bitgo_1.getInternalChainCode)(input.scriptType === 'taprootKeyPathSpend' ? 'p2trMusig2' : input.scriptType);
        return (0, mock_1.mockWalletUnspent)(network, input.value, {
            chain,
            vout: index,
            keys: rootWalletKeys,
            index,
        });
    }
}
/**
 * returns signer and cosigner names for InputScriptType.
 * user and undefined as signer and cosigner respectively for p2shP2pk.
 * user and backup as signer and cosigner respectively for p2trMusig2.
 * user and bitgo as signer and cosigner respectively for other input script types.
 */
function getSigners(inputType) {
    return {
        signerName: 'user',
        cosignerName: inputType === 'p2shP2pk' ? undefined : inputType === 'p2trMusig2' ? 'backup' : 'bitgo',
    };
}
/**
 * signs with first or second signature for single input.
 * p2shP2pk is signed only with first sign.
 */
function signPsbtInput(psbt, input, inputIndex, rootWalletKeys, sign, params) {
    const { signers, deterministic } = params ?? {};
    const { signerName, cosignerName } = signers ? signers : getSigners(input.scriptType);
    if (sign === 'halfsigned') {
        if (input.scriptType === 'p2shP2pk') {
            psbt.signInput(inputIndex, rootWalletKeys[signerName]);
        }
        else {
            psbt.signInputHD(inputIndex, rootWalletKeys[signerName]);
        }
    }
    if (sign === 'fullsigned' && cosignerName && input.scriptType !== 'p2shP2pk') {
        psbt.signInputHD(inputIndex, rootWalletKeys[cosignerName], { deterministic });
    }
}
/**
 * signs with first or second signature for all inputs.
 * p2shP2pk is signed only with first sign.
 */
function signAllPsbtInputs(psbt, inputs, rootWalletKeys, sign, params) {
    const { signers, deterministic } = params ?? {};
    inputs.forEach((input, inputIndex) => {
        signPsbtInput(psbt, input, inputIndex, rootWalletKeys, sign, { signers, deterministic });
    });
}
/**
 * construct psbt for given inputs, outputs, network and root wallet keys.
 */
function constructPsbt(inputs, outputs, network, rootWalletKeys, signStage, params) {
    const { signers, deterministic, skipNonWitnessUtxo } = params ?? {};
    const totalInputAmount = inputs.reduce((sum, input) => sum + input.value, BigInt(0));
    const outputInputAmount = outputs.reduce((sum, output) => sum + output.value, BigInt(0));
    (0, assert_1.ok)(totalInputAmount >= outputInputAmount, 'total output can not exceed total input');
    const psbt = (0, bitgo_1.createPsbtForNetwork)({ network });
    if (params?.addGlobalXPubs) {
        (0, bitgo_1.addXpubsToPsbt)(psbt, rootWalletKeys);
    }
    const unspents = inputs.map((input, i) => toUnspent(input, i, network, rootWalletKeys, { p2shP2pkKey: params?.p2shP2pkKey }));
    unspents.forEach((u, i) => {
        const { signerName, cosignerName } = signers ? signers : getSigners(inputs[i].scriptType);
        if ((0, bitgo_1.isWalletUnspent)(u) && cosignerName) {
            (0, bitgo_1.addWalletUnspentToPsbt)(psbt, u, rootWalletKeys, signerName, cosignerName, { skipNonWitnessUtxo });
        }
        else {
            const { redeemScript } = (0, outputScripts_1.createOutputScriptP2shP2pk)(params?.p2shP2pkKey ?? rootWalletKeys.user.publicKey);
            (0, assert_1.ok)(redeemScript);
            (0, bitgo_1.addReplayProtectionUnspentToPsbt)(psbt, u, redeemScript, { skipNonWitnessUtxo });
        }
    });
    outputs.forEach((output, i) => {
        if ('scriptType' in output) {
            (0, bitgo_1.addWalletOutputToPsbt)(psbt, output.walletKeys ?? rootWalletKeys, output.isInternalAddress ? (0, bitgo_1.getInternalChainCode)(output.scriptType) : (0, bitgo_1.getExternalChainCode)(output.scriptType), i, output.value, { addDerivationInfo: output.walletKeys !== null });
            return;
        }
        else if ('address' in output) {
            const { address, value } = output;
            psbt.addOutput({ script: (0, address_1.toOutputScript)(address, network), value });
            return;
        }
        else if ('opReturn' in output) {
            const { opReturn, value } = output;
            const script = bitcoinjslib.payments.embed({ data: [Buffer.from(opReturn, 'ascii')] }).output;
            (0, assert_1.ok)(script, 'script is required');
            psbt.addOutput({ script, value });
            return;
        }
        else if ('script' in output) {
            const { script, value } = output;
            psbt.addOutput({ script: Buffer.from(script, 'hex'), value });
            return;
        }
        throw new Error('invalid output');
    });
    if (signStage === 'unsigned') {
        return psbt;
    }
    /* use fixed sessionId for deterministic nonce creation */
    const sessionId = Buffer.alloc(32);
    psbt.setAllInputsMusig2NonceHD(rootWalletKeys['user'], { sessionId });
    psbt.setAllInputsMusig2NonceHD(rootWalletKeys['bitgo'], { deterministic });
    signAllPsbtInputs(psbt, inputs, rootWalletKeys, 'halfsigned', { signers, skipNonWitnessUtxo });
    if (signStage === 'fullsigned') {
        signAllPsbtInputs(psbt, inputs, rootWalletKeys, signStage, { signers, deterministic, skipNonWitnessUtxo });
    }
    return psbt;
}
exports.txFormats = ['psbt', 'psbt-lite'];
/**
 * Creates a valid PSBT with as many features as possible.
 *
 * - Inputs:
 *   - All wallet script types that are supported by the network.
 *   - A p2shP2pk input (for replay protection)
 * - Outputs:
 *   - All wallet script types that are supported by the network.
 *   - A p2sh output with derivation info of a different wallet (not in the global psbt xpubs)
 *   - A p2sh output with no derivation info (external output)
 *   - An OP_RETURN output
 */
class AcidTest {
    constructor(network, signStage, txFormat, rootWalletKeys, otherWalletKeys, inputs, outputs) {
        this.network = network;
        this.signStage = signStage;
        this.txFormat = txFormat;
        this.rootWalletKeys = rootWalletKeys;
        this.otherWalletKeys = otherWalletKeys;
        this.inputs = inputs;
        this.outputs = outputs;
    }
    static withConfig(network, signStage, txFormat, suiteConfig) {
        const rootWalletKeys = (0, keys_1.getDefaultWalletKeys)();
        const otherWalletKeys = (0, keys_1.getWalletKeysForSeed)('too many secrets');
        const inputs = exports.inputScriptTypes
            .filter((scriptType) => scriptType === 'taprootKeyPathSpend'
            ? (0, outputScripts_1.isSupportedScriptType)(network, 'p2trMusig2')
            : (0, outputScripts_1.isSupportedScriptType)(network, scriptType))
            .filter((scriptType) => (suiteConfig.includeP2trMusig2ScriptPath ?? true) || scriptType !== 'p2trMusig2')
            .map((scriptType) => ({ scriptType, value: BigInt(2000) }));
        const outputs = exports.outputScriptTypes
            .filter((scriptType) => (0, outputScripts_1.isSupportedScriptType)(network, scriptType))
            .map((scriptType) => ({ scriptType, value: BigInt(900) }));
        // Test other wallet output (with derivation info)
        outputs.push({ scriptType: 'p2sh', value: BigInt(900), walletKeys: otherWalletKeys });
        // Tes non-wallet output
        outputs.push({ scriptType: 'p2sh', value: BigInt(900), walletKeys: null });
        // Test OP_RETURN output
        outputs.push({ opReturn: 'setec astronomy', value: BigInt(0) });
        return new AcidTest(network, signStage, txFormat, rootWalletKeys, otherWalletKeys, inputs, outputs);
    }
    get name() {
        const networkName = (0, networks_1.getNetworkName)(this.network);
        return `${networkName} ${this.signStage} ${this.txFormat}`;
    }
    getReplayProtectionPublicKey() {
        return this.rootWalletKeys.user.publicKey;
    }
    getReplayProtectionOutputScript() {
        const { scriptPubKey } = (0, outputScripts_1.createOutputScriptP2shP2pk)(this.getReplayProtectionPublicKey());
        (0, assert_1.ok)(scriptPubKey);
        return scriptPubKey;
    }
    createPsbt() {
        const psbt = constructPsbt(this.inputs, this.outputs, this.network, this.rootWalletKeys, this.signStage, {
            deterministic: true,
            addGlobalXPubs: true,
        });
        if (this.txFormat === 'psbt-lite') {
            return (0, bitgo_1.clonePsbtWithoutNonWitnessUtxo)(psbt);
        }
        return psbt;
    }
    static suite(suiteConfig = {}) {
        return (0, networks_1.getNetworkList)()
            .filter((network) => (0, networks_1.isMainnet)(network) && network !== networks_1.networks.bitcoinsv)
            .flatMap((network) => exports.signStages.flatMap((signStage) => exports.txFormats.flatMap((txFormat) => AcidTest.withConfig(network, signStage, txFormat, suiteConfig))));
    }
}
exports.AcidTest = AcidTest;
/**
 * Verifies signatures of fully signed tx (with taproot key path support).
 * NOTE: taproot key path tx can only be built and signed with PSBT.
 */
function verifyFullySignedSignatures(tx, unspents, walletKeys, signer, cosigner) {
    const prevOutputs = unspents.map((u) => (0, bitgo_1.toOutput)(u, tx.network));
    return unspents.every((u, index) => {
        if ((0, bitgo_1.parseSignatureScript2Of3)(tx.ins[index]).scriptType === 'taprootKeyPathSpend') {
            const result = (0, bitgo_1.getSignatureVerifications)(tx, index, u.value, undefined, prevOutputs);
            return result.length === 1 && result[0].signature;
        }
        else {
            const result = (0, bitgo_1.verifySignatureWithUnspent)(tx, index, unspents, walletKeys);
            if ((signer === 'user' && cosigner === 'bitgo') || (signer === 'bitgo' && cosigner === 'user')) {
                return result[0] && !result[1] && result[2];
            }
            else if ((signer === 'user' && cosigner === 'backup') || (signer === 'backup' && cosigner === 'user')) {
                return result[0] && result[1] && !result[2];
            }
            else {
                return !result[0] && result[1] && result[2];
            }
        }
    });
}
