// @bitgo/public-types/src/schema/wallet/address.ts
import * as t from "io-ts";
import { date, DateFromISOString } from "io-ts-types";
import { CoinString } from "../coin/coin";

/**
 * A human-readable label for the address
 * @maxLength 250
 * @example Bob's Hot Wallet Address
 */
export const AddressLabelString = t.string;
export type AddressLabelString = t.TypeOf<typeof AddressLabelString>;

/**
 * @maxLength 250
 * @example 2MvrwRYBAuRtPTiZ5MyKg42Ke55W3fZJfZS
 */
export const AddressString = t.string;
export type AddressString = t.TypeOf<typeof AddressString>;

/**
 * Address chain identifier
 */
export const AddressChain = t.union([
  t.literal(0),
  t.literal(1),
  t.literal(10),
  t.literal(11),
  t.literal(20),
  t.literal(21),
  t.literal(30),
  t.literal(31),
  t.literal(40),
  t.literal(41),
]);
export type AddressChain = t.TypeOf<typeof AddressChain>;

export const AddressChainString = t.union([
  t.literal("0"),
  t.literal("1"),
  t.literal("10"),
  t.literal("11"),
  t.literal("20"),
  t.literal("21"),
  t.literal("30"),
  t.literal("31"),
  t.literal("40"),
  t.literal("41"),
]);
export type AddressChainString = t.TypeOf<typeof AddressChainString>;

/**
 * Balance information for an address
 */
export const TrimmedBalance = t.intersection([
  t.type({
    /**
     * String representation of the balance. Guaranteed to not lose precision.
     * @example 500000
     */
    balanceString: t.string,
    /**
     * The total balance of confirmed transactions for the address (in the chain's base unit). Guaranteed to not lose precision.
     * @example 400000
     */
    confirmedBalanceString: t.string,
    /**
     * The total balance of the address (in the chain's base unit) which may be used as inputs for creating new transactions in string representation. Guaranteed to not lose precision.
     * @example 40000
     */
    spendableBalanceString: t.string,
  }),
  t.partial({
    /**
     * The last time a transaction affected the balance of this address
     */
    updated: t.union([DateFromISOString, date, t.unknown]),
    /**
     * @example 50000
     */
    balance: t.number,
    lockedBalanceString: t.string,
    reservedBalanceString: t.string,
    /**
     * The total amount received on this address (in the chain's base unit)
     */
    totalReceived: t.number,
    /**
     * The total amount sent from this address (in the chain's base unit)
     */
    totalSent: t.number,
    transferCount: t.number,
    tokens: t.UnknownRecord,
    nfts: t.UnknownRecord,
    unsupportedNfts: t.UnknownRecord,
    type: t.unknown,
    metadata: t.unknown,
  }),
]);
export type TrimmedBalance = t.TypeOf<typeof TrimmedBalance>;

/**
 * Base wallet address with optional fields
 */
export const BaseWalletAddress = t.partial({
  /**
   * Platform public ID for an address
   * @example 59cd72485007a239fb00282ed480da1f
   * @pattern ^[0-9a-f]{32}$
   */
  id: t.string,
  address: AddressString,
  balance: t.union([TrimmedBalance, t.unknown]),
  /**
   * Properties specific to certain coin types
   */
  // TODO: Add the coin specific schema
  coinSpecific: t.unknown,
  label: AddressLabelString,
  /**
   * @default -1
   */
  lastNonce: t.number,
  token: t.string,
  proof: t.string,
  signature: t.string,
  lastConsolidatedTime: date,
  needsConsolidation: t.boolean,
  tokenConsolidationState: t.unknown,
  tokenAddress: t.unknown,
});
export type BaseWalletAddress = t.TypeOf<typeof BaseWalletAddress>;

/**
 * Lightning wallet address
 */
export const LightningWalletAddress = t.intersection([
  BaseWalletAddress,
  t.type({
    coin: CoinString,
    wallet: t.string,
  }),
]);
export type LightningWalletAddress = t.TypeOf<typeof LightningWalletAddress>;

/**
 * Standard wallet address
 */
export const WalletAddress = t.intersection([
  BaseWalletAddress,
  t.type({
    coin: CoinString,
    wallet: t.string,
    chain: AddressChain,
    index: t.number,
  }),
]);
export type WalletAddress = t.TypeOf<typeof WalletAddress>;

/**
 * Union of wallet address types
 */
export const WalletAddressUnion = t.union([
  WalletAddress,
  LightningWalletAddress,
]);
export type WalletAddressUnion = t.TypeOf<typeof WalletAddressUnion>;

/**
 * Response type for listing wallet addresses
 */
export const AddressQueryResult = t.intersection([
  t.type({
    coin: t.string,
  }),
  t.partial({
    /** Total number of addresses pending on-chain initialization on this wallet **Note:** for wallets with many addresses (100,000 or more), this property may be omitted for performance reasons. If the pending address count is needed for a large wallet, the `includeTotalAddressCount` query parameter can be used to force an address count to be returned with the initial page of addresses. */
    pendingAddressCount: t.number,
    /** Total number of addresses which match the provided query parameters. **Note:** for wallets with many addresses (100,000 or more), this property may be omitted for performance reasons. If the total address count is needed for a large wallet, the `includeTotalAddressCount` query parameter can be used to force an address count to be returned with the initial page of addresses. */
    totalAddressCount: t.number,
    addresses: t.union([t.array(WalletAddress), t.unknown]),
    /** When a result set is truncated, this field returns the id of the last object in the previous batch.  To get the next batch of results, pass this value via the `prevId` query parameter.
     * @format uuid
     * @example 585951a5df8380e0e3063e9f
     */
    nextBatchPrevId: t.string,
    /** Total number of addresses returned in this response */
    count: t.number,
  }),
]);
export type AddressQueryResult = t.TypeOf<typeof AddressQueryResult>;
