// @bitgo/sdk-coin-sol/dist/esm/lib/closeAtaBuilder.js
import { BuildTransactionError, TransactionType } from '@bitgo/sdk-core';
import assert from 'assert';
import { InstructionBuilderTypes } from './constants';
import { Transaction } from './transaction';
import { TransactionBuilder } from './transactionBuilder';
import { validateAddress } from './utils';
export class CloseAtaBuilder extends TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new Transaction(_coinConfig);
    }
    get transactionType() {
        return TransactionType.CloseAssociatedTokenAccount;
    }
    accountAddress(accountAddress) {
        validateAddress(accountAddress, 'accountAddress');
        this._accountAddress = accountAddress;
        return this;
    }
    destinationAddress(destinationAddress) {
        validateAddress(destinationAddress, 'destinationAddress');
        this._destinationAddress = destinationAddress;
        return this;
    }
    authorityAddress(authorityAddress) {
        validateAddress(authorityAddress, 'authorityAddress');
        this._authorityAddress = authorityAddress;
        return this;
    }
    /** @inheritDoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === InstructionBuilderTypes.CloseAssociatedTokenAccount) {
                const ataCloseInstruction = instruction;
                this.accountAddress(ataCloseInstruction.params.accountAddress);
                this.destinationAddress(ataCloseInstruction.params.destinationAddress);
                this.authorityAddress(ataCloseInstruction.params.authorityAddress);
            }
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        assert(this._accountAddress, 'Account Address must be set before building the transaction');
        assert(this._destinationAddress, 'Destination Address must be set before building the transaction');
        assert(this._authorityAddress, 'Authority Address must be set before building the transaction');
        if (this._accountAddress === this._destinationAddress) {
            throw new BuildTransactionError('Account address to close cannot be the same as the destination address');
        }
        const closeAssociatedTokenAccountData = {
            type: InstructionBuilderTypes.CloseAssociatedTokenAccount,
            params: {
                accountAddress: this._accountAddress,
                destinationAddress: this._destinationAddress,
                authorityAddress: this._authorityAddress,
            },
        };
        this._instructionsData = [closeAssociatedTokenAccountData];
        return await super.buildImplementation();
    }
}
