// @bitgo/sdk-coin-sol/dist/esm/lib/constants.js
export const MEMO_PROGRAM_PK = 'MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr';
export const SEED_LENGTH = 32;
export const MAX_MEMO_LENGTH = 130;
export const STAKE_ACCOUNT_RENT_EXEMPT_AMOUNT = 2282880;
export const UNAVAILABLE_TEXT = 'UNAVAILABLE';
/**
 * Maximum over-the-wire size of a Solana transaction (in bytes)
 *
 * Source: https://github.com/anza-xyz/agave/blob/v2.1.13/sdk/packet/src/lib.rs#L27-L29
 *
 * Calculation:
 * - IPv6 minimum MTU: 1280 bytes
 * - IPv6 header: 40 bytes
 * - Fragment/UDP header: 8 bytes
 * - Result: 1280 - 40 - 8 = 1232 bytes
 *
 * This limit is designed to avoid packet fragmentation on typical internet infrastructure.
 * Transactions exceeding this limit will fail to serialize with a RangeError during
 * the encoding of the transaction message.
 *
 * Reference: https://solana.com/docs/core/transactions#transaction-size
 */
export const SOLANA_TRANSACTION_MAX_SIZE = 1232;
export const JITO_STAKE_POOL_ADDRESS = 'Jito4APyf642JPZPx3hGc6WWJ8zPKtRbRs4P815Awbb';
export const JITOSOL_MINT_ADDRESS = 'J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn';
export const JITO_STAKE_POOL_RESERVE_ACCOUNT = 'BgKUXdS29YcHCFrPm5M8oLHiTzZaMDjsebggjoaQ6KFL';
export const JITO_STAKE_POOL_RESERVE_ACCOUNT_TESTNET = 'rrWBQqRqBXYZw3CmPCCcjFxQ2Ds4JFJd7oRQJ997dhz';
export const JITO_MANAGER_FEE_ACCOUNT = 'feeeFLLsam6xZJFc6UQFrHqkvVt4jfmVvi2BRLkUZ4i';
export const JITO_MANAGER_FEE_ACCOUNT_TESTNET = 'DH7tmjoQ5zjqcgfYJU22JqmXhP5EY1tkbYpgVWUS2oNo';
// Sdk instructions, mainly to check decoded types.
export var ValidInstructionTypesEnum;
(function (ValidInstructionTypesEnum) {
    ValidInstructionTypesEnum["AdvanceNonceAccount"] = "AdvanceNonceAccount";
    ValidInstructionTypesEnum["Create"] = "Create";
    ValidInstructionTypesEnum["InitializeNonceAccount"] = "InitializeNonceAccount";
    ValidInstructionTypesEnum["StakingInitialize"] = "Initialize";
    ValidInstructionTypesEnum["StakingDelegate"] = "Delegate";
    ValidInstructionTypesEnum["StakingDeactivate"] = "Deactivate";
    ValidInstructionTypesEnum["StakingWithdraw"] = "Withdraw";
    ValidInstructionTypesEnum["Transfer"] = "Transfer";
    ValidInstructionTypesEnum["TokenTransfer"] = "TokenTransfer";
    ValidInstructionTypesEnum["Memo"] = "Memo";
    ValidInstructionTypesEnum["InitializeAssociatedTokenAccount"] = "InitializeAssociatedTokenAccount";
    ValidInstructionTypesEnum["CloseAssociatedTokenAccount"] = "CloseAssociatedTokenAccount";
    ValidInstructionTypesEnum["Allocate"] = "Allocate";
    ValidInstructionTypesEnum["Assign"] = "Assign";
    ValidInstructionTypesEnum["Split"] = "Split";
    ValidInstructionTypesEnum["Authorize"] = "Authorize";
    ValidInstructionTypesEnum["SetComputeUnitLimit"] = "SetComputeUnitLimit";
    ValidInstructionTypesEnum["SetPriorityFee"] = "SetPriorityFee";
    ValidInstructionTypesEnum["MintTo"] = "MintTo";
    ValidInstructionTypesEnum["Burn"] = "Burn";
    ValidInstructionTypesEnum["DepositSol"] = "DepositSol";
    ValidInstructionTypesEnum["WithdrawStake"] = "WithdrawStake";
    ValidInstructionTypesEnum["Approve"] = "Approve";
    ValidInstructionTypesEnum["CustomInstruction"] = "CustomInstruction";
})(ValidInstructionTypesEnum || (ValidInstructionTypesEnum = {}));
// Internal instructions types
export var InstructionBuilderTypes;
(function (InstructionBuilderTypes) {
    InstructionBuilderTypes["CreateNonceAccount"] = "CreateNonceAccount";
    InstructionBuilderTypes["StakingActivate"] = "Activate";
    InstructionBuilderTypes["StakingDeactivate"] = "Deactivate";
    InstructionBuilderTypes["StakingWithdraw"] = "Withdraw";
    InstructionBuilderTypes["Transfer"] = "Transfer";
    InstructionBuilderTypes["Memo"] = "Memo";
    InstructionBuilderTypes["NonceAdvance"] = "NonceAdvance";
    InstructionBuilderTypes["CreateAssociatedTokenAccount"] = "CreateAssociatedTokenAccount";
    InstructionBuilderTypes["CloseAssociatedTokenAccount"] = "CloseAssociatedTokenAccount";
    InstructionBuilderTypes["TokenTransfer"] = "TokenTransfer";
    InstructionBuilderTypes["StakingAuthorize"] = "Authorize";
    InstructionBuilderTypes["StakingDelegate"] = "Delegate";
    InstructionBuilderTypes["SetComputeUnitLimit"] = "SetComputeUnitLimit";
    InstructionBuilderTypes["SetPriorityFee"] = "SetPriorityFee";
    InstructionBuilderTypes["MintTo"] = "MintTo";
    InstructionBuilderTypes["Burn"] = "Burn";
    InstructionBuilderTypes["CustomInstruction"] = "CustomInstruction";
    InstructionBuilderTypes["VersionedCustomInstruction"] = "VersionedCustomInstruction";
    InstructionBuilderTypes["Approve"] = "Approve";
    InstructionBuilderTypes["WithdrawStake"] = "WithdrawStake";
})(InstructionBuilderTypes || (InstructionBuilderTypes = {}));
export const VALID_SYSTEM_INSTRUCTION_TYPES = [
    ValidInstructionTypesEnum.AdvanceNonceAccount,
    ValidInstructionTypesEnum.Create,
    ValidInstructionTypesEnum.StakingInitialize,
    ValidInstructionTypesEnum.StakingDelegate,
    ValidInstructionTypesEnum.StakingDeactivate,
    ValidInstructionTypesEnum.StakingWithdraw,
    ValidInstructionTypesEnum.Transfer,
    ValidInstructionTypesEnum.InitializeNonceAccount,
    ValidInstructionTypesEnum.Memo,
    ValidInstructionTypesEnum.InitializeAssociatedTokenAccount,
    ValidInstructionTypesEnum.CloseAssociatedTokenAccount,
    ValidInstructionTypesEnum.TokenTransfer,
    ValidInstructionTypesEnum.Allocate,
    ValidInstructionTypesEnum.Assign,
    ValidInstructionTypesEnum.Split,
    ValidInstructionTypesEnum.Authorize,
    ValidInstructionTypesEnum.SetComputeUnitLimit,
    ValidInstructionTypesEnum.SetPriorityFee,
    ValidInstructionTypesEnum.MintTo,
    ValidInstructionTypesEnum.Burn,
    ValidInstructionTypesEnum.Approve,
    ValidInstructionTypesEnum.DepositSol,
    ValidInstructionTypesEnum.WithdrawStake,
    ValidInstructionTypesEnum.CustomInstruction,
];
/** Const to check the order of the Wallet Init instructions when decode */
export const walletInitInstructionIndexes = {
    Create: 0,
    InitializeNonceAccount: 1,
    Memo: 2,
};
/** Const to check the order of the Staking Activate instructions when decode */
export const stakingActivateInstructionsIndexes = {
    Create: 0,
    Initialize: 1,
    Delegate: 2,
    Memo: 3,
};
/** Const to check the order of the Marinade Staking Activate instructions when decode */
export const marinadeStakingActivateInstructionsIndexes = {
    Create: 0,
    Initialize: 1,
    Memo: 2,
};
/** Const to check the order of the Jito Staking Activate instructions when decode */
export const jitoStakingActivateInstructionsIndexes = {
    DepositSol: 0,
};
/** Const to check the order of the Jito Staking Activate instructions when decode */
export const jitoStakingActivateWithATAInstructionsIndexes = {
    InitializeAssociatedTokenAccount: 0,
    DepositSol: 1,
};
/** Const to check the order of the Jito Staking Activate instructions when decode */
export const jitoStakingDeactivateInstructionsIndexes = {
    Approve: 0,
    Create: 1,
    WithdrawStake: 2,
};
/** Const to check the order of the Staking Authorize instructions when decode */
export const stakingAuthorizeInstructionsIndexes = {
    Authorize: 0,
};
/** Const to check the order of the Staking Delegate instructions when decode */
export const stakingDelegateInstructionsIndexes = {
    Delegate: 0,
};
/** Const to check the order of the Staking Deactivate instructions when decode */
export const stakingDeactivateInstructionsIndexes = {
    Deactivate: 0,
    Memo: 1,
};
/** Const to check the order of the Marinade Staking Deactivate instructions when decode */
export const marinadeStakingDeactivateInstructionsIndexes = {
    Transfer: 0,
    Memo: 1,
};
/** Const to check the order of the Partial Staking Deactivate instructions when decoded */
export const stakingPartialDeactivateInstructionsIndexes = {
    Transfer: 0,
    Allocate: 1,
    Assign: 2,
    Split: 3,
    Deactivate: 4,
    Memo: 5,
};
/** Const to check the order of the Staking Withdraw instructions when decode */
export const stakingWithdrawInstructionsIndexes = {
    Withdraw: 0,
    Memo: 1,
};
/** Const to check the order of the ATA init instructions when decode */
export const ataInitInstructionIndexes = {
    InitializeAssociatedTokenAccount: 0,
    Memo: 1,
};
/** Const to check the order of the ATA init instructions when decode */
export const ataCloseInstructionIndexes = {
    CloseAssociatedTokenAccount: 0,
};
export const nonceAdvanceInstruction = 'AdvanceNonceAccount';
export const validInstructionData = '0a00000001000000';
export const validInstructionData2 = '0a00000000000000';
