// @bitgo/sdk-coin-sol/dist/esm/lib/stakingAuthorizeBuilder.js
import { TransactionType } from '@bitgo/sdk-core';
import { TransactionBuilder } from './transactionBuilder';
import { InstructionBuilderTypes } from './constants';
import assert from 'assert';
import { validateAddress } from './utils';
export class StakingAuthorizeBuilder extends TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return TransactionType.StakingAuthorize;
    }
    /** @inheritdoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === InstructionBuilderTypes.StakingAuthorize) {
                const AuthorizeInstruction = instruction;
                this.stakingAddress(AuthorizeInstruction.params.stakingAddress);
                this.newAuthorizedAddress(AuthorizeInstruction.params.newAuthorizeAddress);
                this.oldAuthorizedAddress(AuthorizeInstruction.params.oldAuthorizeAddress);
            }
        }
    }
    /**
     * The address of the staking account.
     *
     * @param {string} stakingAddress public address of the staking account.
     * @returns {StakeBuilder} This staking builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    stakingAddress(stakingAddress) {
        validateAddress(stakingAddress, 'stakingAddress');
        this._stakingAddress = stakingAddress;
        return this;
    }
    /**
     * The address of the new staking account authorization.
     *
     * @param {string} stakingAddress public address of the staking account.
     * @returns {StakeBuilder} This staking builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    newAuthorizedAddress(newAuthorizedAddress) {
        validateAddress(newAuthorizedAddress, 'newAuthorizedAddress');
        this._newAuthorizedAddress = newAuthorizedAddress;
        return this;
    }
    /**
     * The address of the new staking account authorization.
     *
     * @param {string} stakingAddress public address of the staking account.
     * @returns {StakeBuilder} This staking builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    oldAuthorizedAddress(oldAuthorizedAddress) {
        validateAddress(oldAuthorizedAddress, 'oldAuthorizedAddress');
        this._oldAuthorizedAddress = oldAuthorizedAddress;
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        assert(this._stakingAddress, 'Staking Address must be set before building the transaction');
        assert(this._newAuthorizedAddress, 'new authorized Address must be set before building the transaction');
        assert(this._oldAuthorizedAddress, 'old authorized Address must be set before building the transaction');
        const stakingAccountData = {
            type: InstructionBuilderTypes.StakingAuthorize,
            params: {
                stakingAddress: this._stakingAddress,
                newWithdrawAddress: this._newAuthorizedAddress,
                newAuthorizeAddress: this._newAuthorizedAddress,
                oldAuthorizeAddress: this._oldAuthorizedAddress,
            },
        };
        this._instructionsData = [stakingAccountData];
        return await super.buildImplementation();
    }
}
