// @bitgo/sdk-coin-sol/dist/esm/lib/stakingDeactivateBuilder.js
import { SolStakingTypeEnum } from '@bitgo/public-types';
import assert from 'assert';
import { BuildTransactionError, TransactionType } from '@bitgo/sdk-core';
import { InstructionBuilderTypes, STAKE_ACCOUNT_RENT_EXEMPT_AMOUNT } from './constants';
import { TransactionBuilder } from './transactionBuilder';
import { isValidStakingAmount, validateAddress } from './utils';
export class StakingDeactivateBuilder extends TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._stakingType = SolStakingTypeEnum.NATIVE;
    }
    get transactionType() {
        return TransactionType.StakingDeactivate;
    }
    /** @inheritdoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        const stakingAddresses = [];
        for (const instruction of this._instructionsData) {
            if (instruction.type === InstructionBuilderTypes.StakingDeactivate) {
                const deactivateInstruction = instruction;
                // Since _stakingAddresses needs to be populated, it gets special treatment.
                stakingAddresses.push(deactivateInstruction.params.stakingAddress);
                // Marinade staking also cares about sender.
                if (deactivateInstruction.params.stakingType !== SolStakingTypeEnum.MARINADE) {
                    this.sender(deactivateInstruction.params.fromAddress);
                }
                // The other values can just be copied.
                if (deactivateInstruction.params.amount !== undefined) {
                    this.amount(deactivateInstruction.params.amount);
                }
                if (deactivateInstruction.params.unstakingAddress !== undefined) {
                    this.unstakingAddress(deactivateInstruction.params.unstakingAddress);
                }
                if (deactivateInstruction.params.recipients !== undefined) {
                    this.recipients(deactivateInstruction.params.recipients);
                }
                if (deactivateInstruction.params.stakingType !== undefined) {
                    this.stakingType(deactivateInstruction.params.stakingType);
                }
                if (deactivateInstruction.params.extraParams !== undefined) {
                    this.extraParams(deactivateInstruction.params.extraParams);
                }
            }
        }
        if (stakingAddresses.length > 1) {
            this.stakingAddresses(stakingAddresses);
        }
        else {
            if (this._stakingType !== SolStakingTypeEnum.MARINADE) {
                this.stakingAddress(stakingAddresses[0]);
            }
        }
    }
    /**
     * The staking address of the staking account.
     *
     * @param {string} stakingAddress public address of the staking account
     * @returns {StakingDeactivateBuilder} This staking deactivate builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    stakingAddress(stakingAddress) {
        validateAddress(stakingAddress, 'stakingAddress');
        this._stakingAddress = stakingAddress;
        return this;
    }
    /**
     * The staking addresses of the staking account.
     *
     * @param {string[]} stakingAddresses public address of the staking accounts
     * @returns {StakingDeactivateBuilder} This staking deactivate builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    stakingAddresses(stakingAddresses) {
        for (const stakingAddress of stakingAddresses) {
            validateAddress(stakingAddress, 'stakingAddress');
        }
        this._stakingAddresses = stakingAddresses;
        return this;
    }
    /**
     * Optional amount to unstake expressed in Lamports, 1 SOL = 1_000_000_000 lamports, to be used
     * when partially unstaking. If not given then the entire staked amount will be unstaked.
     *
     * @param {string} amount The partial amount to unstake, expressed in Lamports.
     * @returns {StakingDeactivateBuilder} This staking builder.
     *
     * @see https://docs.solana.com/cli/delegate-stake#split-stake
     */
    amount(amount) {
        if (!isValidStakingAmount(amount)) {
            throw new BuildTransactionError('If given, amount cannot be zero or less');
        }
        this._amount = amount;
        return this;
    }
    /**
     * Setter to set the recipients object
     *
     * @param recipients RecipientEntry[] - The recipients object
     * @returns {StakingDeactivateBuilder} This staking builder.
     */
    recipients(recipients) {
        this._recipients = recipients;
        return this;
    }
    /**
     * When partially unstaking move the amount to unstake to this account and initiate the
     * unstake process. The original stake account will continue staking.
     *
     * @param {string} unstakingAddress An account used to unstake a partial amount.
     * @returns {StakingDeactivateBuilder} This staking builder.
     *
     * @see https://docs.solana.com/cli/delegate-stake#split-stake
     */
    unstakingAddress(unstakingAddress) {
        validateAddress(unstakingAddress, 'unstakingAddress');
        this._unstakingAddress = unstakingAddress;
        return this;
    }
    /**
     * Set staking type.
     *
     * @param {SolStakingType} stakingType a staking type.
     * @returns {StakingDeactivateBuilder} This staking builder.
     */
    stakingType(stakingType) {
        this._stakingType = stakingType;
        return this;
    }
    /**
     * Set parameters specific to a staking type.
     *
     * @param {StakingDeactivateExtraParams} extraParams parameters specific to a staking type.
     * @returns {StakingDeactivateBuilder} This staking builder.
     */
    extraParams(extraParams) {
        this._extraParams = extraParams;
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        assert(this._sender, 'Sender must be set before building the transaction');
        if (this._stakingAddresses && this._stakingAddresses.length > 0) {
            this._instructionsData = [];
            for (const stakingAddress of this._stakingAddresses) {
                const stakingDeactivateData = {
                    type: InstructionBuilderTypes.StakingDeactivate,
                    params: {
                        fromAddress: this._sender,
                        stakingAddress: stakingAddress,
                        stakingType: SolStakingTypeEnum.NATIVE,
                    },
                };
                this._instructionsData.push(stakingDeactivateData);
            }
        }
        else {
            if (this._stakingType === SolStakingTypeEnum.NATIVE) {
                // we don't need stakingAddress in marinade staking deactivate txn
                assert(this._stakingAddress, 'Staking address must be set before building the transaction');
            }
            if (this._sender === this._stakingAddress) {
                throw new BuildTransactionError('Sender address cannot be the same as the Staking address');
            }
            if (this._stakingType === SolStakingTypeEnum.NATIVE && this._amount) {
                assert(this._unstakingAddress, 'When partially unstaking the unstaking address must be set before building the transaction');
            }
            this._instructionsData = [];
            if (this._stakingType === SolStakingTypeEnum.NATIVE && this._unstakingAddress) {
                assert(this._amount, 'If an unstaking address is given then a partial amount to unstake must also be set before building the transaction');
                const stakingFundUnstakeAddress = {
                    type: InstructionBuilderTypes.Transfer,
                    params: {
                        fromAddress: this._sender,
                        amount: STAKE_ACCOUNT_RENT_EXEMPT_AMOUNT.toString(),
                        toAddress: this._unstakingAddress,
                    },
                };
                this._instructionsData.push(stakingFundUnstakeAddress);
            }
            const stakingDeactivateData = {
                type: InstructionBuilderTypes.StakingDeactivate,
                params: {
                    fromAddress: this._sender,
                    stakingAddress: this._stakingAddress,
                    amount: this._amount,
                    unstakingAddress: this._unstakingAddress,
                    recipients: this._recipients,
                    stakingType: this._stakingType,
                    extraParams: this._extraParams,
                },
            };
            this._instructionsData.push(stakingDeactivateData);
        }
        return await super.buildImplementation();
    }
}
