// @bitgo/sdk-coin-sol/dist/esm/lib/transactionBuilder.js
import { BaseTransactionBuilder, BuildTransactionError, SigningError, } from '@bitgo/sdk-core';
import { Transaction } from './transaction';
import { MessageV0, PublicKey, Transaction as SolTransaction, VersionedTransaction, } from '@solana/web3.js';
import base58 from 'bs58';
import { isValidAddress, isValidAmount, isValidBlockId, isValidMemo, validateAddress, validateRawTransaction, } from './utils';
import { KeyPair } from '.';
import { InstructionBuilderTypes } from './constants';
import { solInstructionFactory } from './solInstructionFactory';
import assert from 'assert';
import { instructionParamsFactory } from './instructionParamsFactory';
export class TransactionBuilder extends BaseTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._signatures = [];
        this._instructionsData = [];
        this._signers = [];
        this.transaction = new Transaction(_coinConfig);
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        this._transaction = tx;
        const txData = tx.toJson();
        const filteredTransferInstructionsData = txData.instructionsData.filter((data) => data.type === InstructionBuilderTypes.Transfer);
        let sender;
        if (filteredTransferInstructionsData.length > 0) {
            const transferInstructionsData = filteredTransferInstructionsData[0];
            sender = transferInstructionsData.params.fromAddress;
        }
        else {
            sender = txData.feePayer;
        }
        this.sender(sender);
        this.feePayer(txData.feePayer);
        this.nonce(txData.nonce, txData.durableNonce);
        this._instructionsData = instructionParamsFactory(tx.type, tx.solTransaction.instructions, this._coinConfig.name, txData.instructionsData, tx.useTokenAddressTokenName);
        // Parse priority fee instruction data
        const filteredPriorityFeeInstructionsData = txData.instructionsData.filter((data) => data.type === InstructionBuilderTypes.SetPriorityFee);
        for (const instruction of this._instructionsData) {
            if (instruction.type === InstructionBuilderTypes.Memo) {
                const memoInstruction = instruction;
                this.memo(memoInstruction.params.memo);
            }
            if (instruction.type === InstructionBuilderTypes.NonceAdvance) {
                const advanceNonceInstruction = instruction;
                this.nonce(txData.nonce, advanceNonceInstruction.params);
            }
            // If prio fee instruction exists, set the priority fee variable
            if (instruction.type === InstructionBuilderTypes.SetPriorityFee) {
                const priorityFeeInstructionsData = filteredPriorityFeeInstructionsData[0];
                this.setPriorityFee({ amount: Number(priorityFeeInstructionsData.params.fee) });
            }
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = new Transaction(this._coinConfig);
        this.validateRawTransaction(rawTransaction);
        tx.fromRawTransaction(rawTransaction);
        this.initBuilder(tx);
        return this.transaction;
    }
    /** @inheritdoc */
    async buildImplementation() {
        const builtTransaction = this.buildSolTransaction();
        if (builtTransaction instanceof VersionedTransaction) {
            this.transaction.versionedTransaction = builtTransaction;
        }
        else {
            this.transaction.solTransaction = builtTransaction;
        }
        this.transaction.setTransactionType(this.transactionType);
        this.transaction.setInstructionsData(this._instructionsData);
        this.transaction.loadInputsAndOutputs();
        this._transaction.tokenAccountRentExemptAmount = this._tokenAccountRentExemptAmount;
        return this.transaction;
    }
    /**
     * Builds the solana transaction.
     */
    buildSolTransaction() {
        assert(this._sender, new BuildTransactionError('sender is required before building'));
        assert(this._recentBlockhash, new BuildTransactionError('recent blockhash is required before building'));
        // Check if we should build as VersionedTransaction
        if (this._transaction.isVersionedTransaction()) {
            return this.buildVersionedTransaction();
        }
        else {
            return this.buildLegacyTransaction();
        }
    }
    /**
     * Builds a legacy Solana transaction.
     */
    buildLegacyTransaction() {
        const tx = new SolTransaction();
        if (this._transaction?.solTransaction?.signatures) {
            tx.signatures = this._transaction?.solTransaction?.signatures;
        }
        tx.feePayer = this._feePayer ? new PublicKey(this._feePayer) : new PublicKey(this._sender);
        if (this._nonceInfo) {
            tx.nonceInfo = {
                nonce: this._recentBlockhash,
                nonceInstruction: solInstructionFactory(this._nonceInfo)[0],
            };
        }
        else {
            tx.recentBlockhash = this._recentBlockhash;
        }
        for (const instruction of this._instructionsData) {
            tx.add(...solInstructionFactory(instruction));
        }
        if (this._memo) {
            const memoData = {
                type: InstructionBuilderTypes.Memo,
                params: {
                    memo: this._memo,
                },
            };
            this._instructionsData.push(memoData);
            tx.add(...solInstructionFactory(memoData));
        }
        this._transaction.lamportsPerSignature = this._lamportsPerSignature;
        for (const signer of this._signers) {
            const publicKey = new PublicKey(signer.getKeys().pub);
            const secretKey = signer.getKeys(true).prv;
            assert(secretKey instanceof Uint8Array);
            tx.partialSign({ publicKey, secretKey });
        }
        for (const signature of this._signatures) {
            const solPublicKey = new PublicKey(signature.publicKey.pub);
            tx.addSignature(solPublicKey, signature.signature);
        }
        return tx;
    }
    /**
     * Builds a VersionedTransaction.
     *
     * @returns {VersionedTransaction} The built versioned transaction
     */
    buildVersionedTransaction() {
        const versionedTxData = this._transaction.getVersionedTransactionData();
        if (!versionedTxData) {
            throw new BuildTransactionError('Missing VersionedTransactionData');
        }
        const versionedTx = this.buildFromVersionedData(versionedTxData);
        this._transaction.lamportsPerSignature = this._lamportsPerSignature;
        for (const signer of this._signers) {
            const publicKey = new PublicKey(signer.getKeys().pub);
            const secretKey = signer.getKeys(true).prv;
            assert(secretKey instanceof Uint8Array);
            versionedTx.sign([{ publicKey, secretKey }]);
        }
        for (const signature of this._signatures) {
            const solPublicKey = new PublicKey(signature.publicKey.pub);
            versionedTx.addSignature(solPublicKey, signature.signature);
        }
        return versionedTx;
    }
    /**
     * Build a VersionedTransaction from deconstructed VersionedTransactionData
     * @param data The versioned transaction data
     * @returns The built versioned transaction
     */
    buildFromVersionedData(data) {
        const staticAccountKeys = data.staticAccountKeys.map((key) => new PublicKey(key));
        const addressTableLookups = data.addressLookupTables.map((alt) => ({
            accountKey: new PublicKey(alt.accountKey),
            writableIndexes: alt.writableIndexes,
            readonlyIndexes: alt.readonlyIndexes,
        }));
        const compiledInstructions = data.versionedInstructions.map((instruction) => ({
            programIdIndex: instruction.programIdIndex,
            accountKeyIndexes: instruction.accountKeyIndexes,
            data: Buffer.from(base58.decode(instruction.data)),
        }));
        const recentBlockhash = data.recentBlockhash || this._recentBlockhash;
        if (!recentBlockhash) {
            throw new BuildTransactionError('Missing recent blockhash for VersionedTransaction');
        }
        if (!this._sender) {
            throw new BuildTransactionError('Missing sender (fee payer) for VersionedTransaction');
        }
        const messageV0 = new MessageV0({
            header: data.messageHeader,
            staticAccountKeys,
            recentBlockhash: recentBlockhash,
            compiledInstructions,
            addressTableLookups,
        });
        return new VersionedTransaction(messageV0);
    }
    // region Getters and Setters
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    /** @inheritdoc */
    set transaction(transaction) {
        this._transaction = transaction;
    }
    /** @inheritdoc */
    signImplementation(key) {
        this.validateKey(key);
        this.checkDuplicatedSigner(key);
        const prv = key.key;
        const signer = new KeyPair({ prv: prv });
        this._signers.push(signer);
        return this._transaction;
    }
    /** @inheritDoc */
    addSignature(publicKey, signature) {
        this._signatures.push({ publicKey, signature });
    }
    /**
     * Sets the sender of this transaction.
     * This account will be responsible for paying transaction fees.
     *
     * @param {string} senderAddress the account that is sending this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    sender(senderAddress) {
        validateAddress(senderAddress, 'sender');
        this._sender = senderAddress;
        return this;
    }
    /**
     * Set the transaction nonce
     * Requires both optional params in order to use the durable nonce
     *
     * @param {Blockhash} blockHash The latest blockHash
     * @param {DurableNonceParams} [durableNonceParams] An object containing the walletNonceAddress and the authWalletAddress (required for durable nonce)
     * @returns {TransactionBuilder} This transaction builder
     */
    nonce(blockHash, durableNonceParams) {
        if (!blockHash || !isValidBlockId(blockHash)) {
            throw new BuildTransactionError('Invalid or missing blockHash, got: ' + blockHash);
        }
        if (durableNonceParams) {
            validateAddress(durableNonceParams.walletNonceAddress, 'walletNonceAddress');
            validateAddress(durableNonceParams.authWalletAddress, 'authWalletAddress');
            if (durableNonceParams.walletNonceAddress === durableNonceParams.authWalletAddress) {
                throw new BuildTransactionError('Invalid params: walletNonceAddress cannot be equal to authWalletAddress');
            }
            this._nonceInfo = {
                type: InstructionBuilderTypes.NonceAdvance,
                params: durableNonceParams,
            };
        }
        this._recentBlockhash = blockHash;
        return this;
    }
    /**
     *  Set the memo
     *
     * @param {string} memo
     * @returns {TransactionBuilder} This transaction builder
     */
    memo(memo) {
        this.validateMemo(memo);
        this._memo = memo;
        return this;
    }
    fee(feeOptions) {
        this._lamportsPerSignature = Number(feeOptions.amount);
        return this;
    }
    setPriorityFee(feeOptions) {
        this._priorityFee = Number(feeOptions.amount);
        return this;
    }
    feePayer(feePayer) {
        this._feePayer = feePayer;
        return this;
    }
    /**
     * Used to set the minimum rent exempt amount for an ATA
     *
     * @param tokenAccountRentExemptAmount minimum rent exempt amount in lamports
     */
    associatedTokenAccountRent(tokenAccountRentExemptAmount) {
        this.validateRentExemptAmount(tokenAccountRentExemptAmount);
        this._tokenAccountRentExemptAmount = tokenAccountRentExemptAmount;
        return this;
    }
    validateRentExemptAmount(tokenAccountRentExemptAmount) {
        // _tokenAccountRentExemptAmount is allowed to be undefined or a valid amount if it's defined
        if (tokenAccountRentExemptAmount && !isValidAmount(tokenAccountRentExemptAmount)) {
            throw new BuildTransactionError('Invalid tokenAccountRentExemptAmount, got: ' + tokenAccountRentExemptAmount);
        }
    }
    // endregion
    // region Validators
    /** @inheritdoc */
    validateAddress(address, addressFormat) {
        if (!isValidAddress(address.address)) {
            throw new BuildTransactionError('Invalid address ' + address.address);
        }
    }
    /** @inheritdoc */
    validateKey(key) {
        let keyPair;
        try {
            keyPair = new KeyPair({ prv: key.key });
        }
        catch {
            throw new BuildTransactionError('Invalid key');
        }
        if (!keyPair.getKeys().prv) {
            throw new BuildTransactionError('Invalid key');
        }
    }
    /** @inheritdoc */
    validateRawTransaction(rawTransaction) {
        validateRawTransaction(rawTransaction);
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        this.validateSender();
        this.validateNonce();
        this.validateRentExemptAmount(this._tokenAccountRentExemptAmount);
    }
    /** @inheritdoc */
    validateValue(value) {
        if (value.isLessThan(0)) {
            throw new BuildTransactionError('Value cannot be less than zero');
        }
    }
    /** Validates the memo
     *
     * @param {string} memo - the memo as string
     */
    validateMemo(memo) {
        if (!memo) {
            throw new BuildTransactionError('Invalid memo, got: ' + memo);
        }
        if (!isValidMemo(memo)) {
            throw new BuildTransactionError('Memo is too long');
        }
    }
    /**
     * Validates that the given key is not already in this._signers
     *
     * @param {BaseKey} key - The key to check
     */
    checkDuplicatedSigner(key) {
        this._signers.forEach((kp) => {
            if (kp.getKeys().prv === key.key) {
                throw new SigningError('Duplicated signer: ' + key.key);
            }
        });
    }
    /**
     * Validates that the sender field is defined
     */
    validateSender() {
        if (this._sender === undefined) {
            throw new BuildTransactionError('Invalid transaction: missing sender');
        }
    }
    /**
     * Validates that the nonce field is defined
     */
    validateNonce() {
        if (this._recentBlockhash === undefined) {
            throw new BuildTransactionError('Invalid transaction: missing nonce blockhash');
        }
    }
}
