// @bitgo/sdk-coin-sol/dist/esm/lib/walletInitializationBuilder.js
import assert from 'assert';
import { BuildTransactionError, TransactionType } from '@bitgo/sdk-core';
import { TransactionBuilder } from './transactionBuilder';
import { isValidAmount, validateAddress } from './utils';
import { InstructionBuilderTypes } from './constants';
export class WalletInitializationBuilder extends TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return TransactionType.WalletInitialization;
    }
    /** @inheritDoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === InstructionBuilderTypes.CreateNonceAccount) {
                const walletInitInstruction = instruction;
                this.address(walletInitInstruction.params.nonceAddress);
                this.amount(walletInitInstruction.params.amount);
                this.sender(walletInitInstruction.params.authAddress);
            }
        }
    }
    /**
     * Sets the amount to fund the nonce account
     *
     * @param amount amount in lamports to fund the nonce account
     */
    amount(amount) {
        if (!amount || !isValidAmount(amount)) {
            throw new BuildTransactionError('Invalid or missing amount, got: ' + amount);
        }
        this._amount = amount;
        return this;
    }
    /**
     * Sets the address for the nonce account
     * @param nonceAddress address of the new nonce account
     */
    address(nonceAddress) {
        validateAddress(nonceAddress, 'nonceAddress');
        this._nonceAddress = nonceAddress;
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        assert(this._sender, 'Sender must be set before building the transaction');
        assert(this._amount, 'Amount must be set before building the transaction');
        assert(this._nonceAddress, 'Nonce Address must be set before building the transaction');
        const walletInitData = {
            type: InstructionBuilderTypes.CreateNonceAccount,
            params: {
                fromAddress: this._sender,
                nonceAddress: this._nonceAddress,
                authAddress: this._sender,
                amount: this._amount,
            },
        };
        this._instructionsData = [walletInitData];
        return await super.buildImplementation();
    }
}
