// @bitgo/sdk-coin-near/dist/test/unit/transactionBuilder/stakingActivateBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const testData = __importStar(require("../../resources/near"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const sdk_core_1 = require("@bitgo/sdk-core");
const near_1 = require("../../resources/near");
describe('Near Staking Activate Builder', () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('tnear');
    const gas = '125000000000000';
    const amount = '1000000'; // amount to be staked in yoctos
    describe('Succeed', () => {
        it('build a create and delegate staking signed tx', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder
                .amount(amount)
                .gas(gas)
                .sender(testData.accounts.account1.address, testData.accounts.account1.publicKey)
                .receiverId(near_1.validatorContractAddress)
                .recentBlockHash(testData.blockHash.block1)
                .nonce(BigInt(1));
            txBuilder.sign({ key: testData.accounts.account1.secretKey });
            const tx = await txBuilder.build();
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.accounts.account1.address,
                value: amount,
                coin: 'tnear',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: near_1.validatorContractAddress,
                value: amount,
                coin: 'tnear',
            });
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingActivate);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(rawTx, testData.rawTx.stakingActivate.signed);
            const txJson = tx.toJson();
            txJson.should.have.properties(['id', 'signerId', 'publicKey', 'nonce', 'actions', 'signature']);
            txJson.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
            txJson.signerId.should.equal(testData.accounts.account1.address);
            txJson.publicKey.should.equal(testData.accounts.account1.publicKeyBase58);
            txJson.nonce.should.equal(BigInt(1));
            txJson.receiverId.should.equal('lavenderfive.pool.f863973.m0');
            txJson.actions.should.deepEqual([
                {
                    functionCall: {
                        methodName: 'deposit_and_stake',
                        args: {},
                        gas: '125000000000000',
                        deposit: '1000000',
                    },
                },
            ]);
        });
        it('build a create and delegate staking unsigned tx', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder
                .amount(amount)
                .gas(gas)
                .sender(testData.accounts.account1.address, testData.accounts.account1.publicKey)
                .receiverId(near_1.validatorContractAddress)
                .recentBlockHash(testData.blockHash.block1)
                .nonce(BigInt(1));
            const tx = await txBuilder.build();
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.accounts.account1.address,
                value: amount,
                coin: 'tnear',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: near_1.validatorContractAddress,
                value: amount,
                coin: 'tnear',
            });
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingActivate);
            const rawTx = tx.toBroadcastFormat();
            should_1.default.equal(rawTx, testData.rawTx.stakingActivate.unsigned);
        });
        it('should fail to create and delegate staking with missing gas', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder
                .amount(amount)
                .sender(testData.accounts.account1.address, testData.accounts.account1.publicKey)
                .receiverId(near_1.validatorContractAddress)
                .recentBlockHash(testData.blockHash.block1)
                .nonce(BigInt(1));
            await txBuilder.build().should.be.rejectedWith('gas is required before building staking activate');
        });
        it('should fail to create and delegate staking with missing amount', async () => {
            const txBuilder = factory.getStakingActivateBuilder();
            txBuilder
                .gas(gas)
                .sender(testData.accounts.account1.address, testData.accounts.account1.publicKey)
                .receiverId(near_1.validatorContractAddress)
                .recentBlockHash(testData.blockHash.block1)
                .nonce(BigInt(1));
            await txBuilder.build().should.be.rejectedWith('amount is required before building staking activate');
        });
        it('build from an unsigned staking activate', async () => {
            const txBuilder = factory.from(testData.rawTx.stakingActivate.unsigned);
            const tx = await txBuilder.build();
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.accounts.account1.address,
                value: '1000000',
                coin: 'tnear',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: 'lavenderfive.pool.f863973.m0',
                value: '1000000',
                coin: 'tnear',
            });
            tx.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
        });
        it('build from an signed staking activate', async () => {
            const txBuilder = factory.from(testData.rawTx.stakingActivate.signed);
            const tx = await txBuilder.build();
            tx.inputs.length.should.equal(1);
            tx.inputs[0].should.deepEqual({
                address: testData.accounts.account1.address,
                value: '1000000',
                coin: 'tnear',
            });
            tx.outputs.length.should.equal(1);
            tx.outputs[0].should.deepEqual({
                address: 'lavenderfive.pool.f863973.m0',
                value: '1000000',
                coin: 'tnear',
            });
            tx.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
            const txJson = tx.toJson();
            txJson.should.have.properties(['id', 'signerId', 'publicKey', 'nonce', 'actions', 'signature']);
            txJson.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
            txJson.signerId.should.equal(testData.accounts.account1.address);
            txJson.publicKey.should.equal(testData.accounts.account1.publicKeyBase58);
            txJson.nonce.should.equal(BigInt(1));
            txJson.receiverId.should.equal('lavenderfive.pool.f863973.m0');
            txJson.actions.should.deepEqual([
                {
                    functionCall: {
                        methodName: 'deposit_and_stake',
                        args: {},
                        gas: '125000000000000',
                        deposit: '1000000',
                    },
                },
            ]);
        });
    });
});
