// @bitgo/sdk-coin-baby/dist/src/lib/BabylonTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BabylonTransaction = void 0;
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
class BabylonTransaction extends abstract_cosmos_1.CosmosTransaction {
    constructor(_coinConfig, _utils) {
        super(_coinConfig, _utils);
    }
    explainTransactionInternal(json, explanationResult) {
        let outputs;
        let outputAmount;
        switch (json.type) {
            case sdk_core_1.TransactionType.CustomTx:
                explanationResult.type = sdk_core_1.TransactionType.CustomTx;
                outputAmount = BigInt(0);
                outputs = json.sendMessages.flatMap((message) => {
                    const value = message.value;
                    switch (value._kind) {
                        case 'CreateBtcDelegation':
                            return [];
                        case 'WithdrawReward':
                            return [
                                {
                                    address: value.address,
                                    amount: constants_1.UNAVAILABLE_TEXT,
                                },
                            ];
                        default:
                            throw new sdk_core_1.InvalidTransactionError(`Unsupported BabylonSpecificMessages message`);
                    }
                });
                break;
            default:
                return super.explainTransactionInternal(json, explanationResult);
        }
        if (json.memo) {
            outputs.forEach((output) => {
                output.memo = json.memo;
            });
        }
        return {
            ...explanationResult,
            outputAmount: outputAmount?.toString(),
            outputs,
        };
    }
    loadInputsAndOutputs() {
        if (this.type === undefined || !this.cosmosLikeTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Transaction type or cosmosLikeTransaction is not set');
        }
        const outputs = [];
        const inputs = [];
        switch (this.type) {
            case sdk_core_1.TransactionType.CustomTx:
                this.cosmosLikeTransaction.sendMessages.forEach((message) => {
                    const value = message.value;
                    switch (value._kind) {
                        case 'CreateBtcDelegation':
                            break;
                        case 'WithdrawReward':
                            inputs.push({
                                address: value.address,
                                value: constants_1.UNAVAILABLE_TEXT,
                                coin: this._coinConfig.name,
                            });
                            outputs.push({
                                address: value.address,
                                value: constants_1.UNAVAILABLE_TEXT,
                                coin: this._coinConfig.name,
                            });
                            break;
                        default:
                            throw new sdk_core_1.InvalidTransactionError(`Unsupported BabylonSpecificMessages message`);
                    }
                });
                break;
            default:
                return super.loadInputsAndOutputs();
        }
        this._inputs = inputs;
        this._outputs = outputs;
    }
}
exports.BabylonTransaction = BabylonTransaction;
