// @bitgo/sdk-coin-baby/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withdrawRewardMsgTypeUrl = exports.createBTCDelegationMsgTypeUrl = exports.wrappedBeginRedelegateTypeUrl = exports.wrappedUndelegateMsgTypeUrl = exports.wrappedDelegateMsgTypeUrl = exports.UNAVAILABLE_TEXT = exports.GAS_LIMIT = exports.GAS_AMOUNT = exports.ADDRESS_PREFIX = exports.contractAddressRegex = exports.validatorAddressRegex = exports.accountAddressRegex = exports.validDenoms = void 0;
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const cosmosUtils = new abstract_cosmos_1.CosmosUtils();
exports.validDenoms = ['baby', 'tbaby', 'ubbn', ...cosmosUtils.getTokenDenomsUsingCoinFamily('baby')];
exports.accountAddressRegex = /^(bbn)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.validatorAddressRegex = /^(bbnvaloper)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.contractAddressRegex = /^(bbn)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]+)$/;
exports.ADDRESS_PREFIX = 'bbn';
exports.GAS_AMOUNT = '7000';
exports.GAS_LIMIT = 200000;
exports.UNAVAILABLE_TEXT = 'UNAVAILABLE';
exports.wrappedDelegateMsgTypeUrl = '/babylon.epoching.v1.MsgWrappedDelegate';
exports.wrappedUndelegateMsgTypeUrl = '/babylon.epoching.v1.MsgWrappedUndelegate';
exports.wrappedBeginRedelegateTypeUrl = '/babylon.epoching.v1.MsgWrappedBeginRedelegate';
exports.createBTCDelegationMsgTypeUrl = '/babylon.btcstaking.v1.MsgCreateBTCDelegation';
exports.withdrawRewardMsgTypeUrl = '/babylon.incentive.MsgWithdrawReward';
