// @bitgo/sdk-coin-baby/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const babylon_proto_ts_1 = require("@babylonlabs-io/babylon-proto-ts");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants = __importStar(require("./constants"));
class Utils extends abstract_cosmos_1.CosmosUtils {
    constructor() {
        super();
        this.babylonMessageKindToTypeUrl = {
            CreateBtcDelegation: constants.createBTCDelegationMsgTypeUrl,
            WithdrawReward: constants.withdrawRewardMsgTypeUrl,
        };
        this.babylonMessageTypeUrlToKind = Object.fromEntries(Object.entries(this.babylonMessageKindToTypeUrl).map(([key, value]) => [value, key]));
        this.wrappedMsgTypeUrls = new Set([
            constants.wrappedDelegateMsgTypeUrl,
            constants.wrappedUndelegateMsgTypeUrl,
            constants.wrappedBeginRedelegateTypeUrl,
        ]);
        this.customMsgTypeUrls = new Set([constants.createBTCDelegationMsgTypeUrl]);
        this.registry.register(constants.wrappedDelegateMsgTypeUrl, babylon_proto_ts_1.epochingtx.MsgWrappedDelegate);
        this.registry.register(constants.wrappedUndelegateMsgTypeUrl, babylon_proto_ts_1.epochingtx.MsgWrappedUndelegate);
        this.registry.register(constants.wrappedBeginRedelegateTypeUrl, babylon_proto_ts_1.epochingtx.MsgWrappedBeginRedelegate);
        this.registry.register(constants.createBTCDelegationMsgTypeUrl, babylon_proto_ts_1.btcstakingtx.MsgCreateBTCDelegation);
        this.registry.register(constants.withdrawRewardMsgTypeUrl, babylon_proto_ts_1.incentivetx.MsgWithdrawReward);
    }
    /** @inheritdoc */
    getDelegateOrUndelegateMessageDataFromDecodedTx(decodedTx) {
        return decodedTx.body.messages.map((message) => {
            const value = this.registry.decode(message).msg;
            return {
                typeUrl: message.typeUrl,
                value: {
                    delegatorAddress: value.delegatorAddress,
                    validatorAddress: value.validatorAddress,
                    amount: value.amount,
                },
            };
        });
    }
    /** @inheritdoc */
    getRedelegateMessageDataFromDecodedTx(decodedTx) {
        return decodedTx.body.messages.map((message) => {
            const value = this.registry.decode(message).msg;
            return {
                typeUrl: message.typeUrl,
                value: {
                    delegatorAddress: value.delegatorAddress,
                    validatorSrcAddress: value.validatorSrcAddress,
                    validatorDstAddress: value.validatorDstAddress,
                    amount: value.amount,
                },
            };
        });
    }
    /** @inheritdoc */
    getCustomMessageDataFromDecodedTx(decodedTx) {
        return decodedTx.body.messages.map((message) => {
            const value = this.registry.decode(message);
            return {
                typeUrl: message.typeUrl,
                value: {
                    _kind: this.babylonMessageTypeUrlToKind[message.typeUrl],
                    ...value,
                },
            };
        });
    }
    /** @inheritdoc */
    getTransactionTypeFromTypeUrl(typeUrl) {
        switch (typeUrl) {
            case constants.wrappedDelegateMsgTypeUrl:
                return sdk_core_1.TransactionType.StakingActivate;
            case constants.wrappedUndelegateMsgTypeUrl:
                return sdk_core_1.TransactionType.StakingDeactivate;
            case constants.wrappedBeginRedelegateTypeUrl:
                return sdk_core_1.TransactionType.StakingRedelegate;
            case constants.createBTCDelegationMsgTypeUrl:
            case constants.withdrawRewardMsgTypeUrl:
                return sdk_core_1.TransactionType.CustomTx;
            default:
                return super.getTransactionTypeFromTypeUrl(typeUrl);
        }
    }
    /** @inheritdoc */
    getSendMessagesForEncodingTx(cosmosLikeTransaction) {
        return cosmosLikeTransaction.sendMessages.map(({ typeUrl, value }) => {
            let valueToEncode = value;
            if (this.wrappedMsgTypeUrls.has(typeUrl)) {
                valueToEncode = { msg: value };
            }
            else if (this.customMsgTypeUrls.has(typeUrl)) {
                // eslint-disable-next-line @typescript-eslint/no-unused-vars
                const { _kind, ...rest } = value;
                valueToEncode = rest;
            }
            return { typeUrl, value: valueToEncode };
        });
    }
    /** @inheritdoc */
    validateCustomMessage(customMessage) {
        switch (customMessage._kind) {
            case 'CreateBtcDelegation':
                this.validateCreateBtcDelegationMessage(customMessage);
                break;
            case 'WithdrawReward':
                this.validateWithdrawRewardMessage(customMessage);
                break;
            default:
                throw new sdk_core_1.InvalidTransactionError(`Unsupported BabylonSpecificMessages message`);
        }
    }
    validateCreateBtcDelegationMessage(createBtcDelegationMessage) {
        if (createBtcDelegationMessage._kind !== 'CreateBtcDelegation') {
            throw new sdk_core_1.InvalidTransactionError(`Invalid CreateBtcDelegationMessage kind: ${createBtcDelegationMessage._kind}`);
        }
        // TODO: check the other fields more thoroughly
        this.isObjPropertyNull(createBtcDelegationMessage, [
            'stakerAddr',
            // 'pop',
            'btcPk',
            'fpBtcPkList',
            'stakingTime',
            'stakingValue',
            'stakingTx',
            // 'stakingTxInclusionProof',
            'slashingTx',
            'delegatorSlashingSig',
            'unbondingTime',
            'unbondingTx',
            'unbondingValue',
            'unbondingSlashingTx',
            'delegatorUnbondingSlashingSig',
        ]);
        if (createBtcDelegationMessage.pop) {
            this.isObjPropertyNull(createBtcDelegationMessage.pop, ['btcSigType', 'btcSig']);
        }
        if (createBtcDelegationMessage.stakingTxInclusionProof) {
            this.isObjPropertyNull(createBtcDelegationMessage.stakingTxInclusionProof, ['key', 'proof']);
            if (createBtcDelegationMessage.stakingTxInclusionProof.key) {
                this.isObjPropertyNull(createBtcDelegationMessage.stakingTxInclusionProof.key, ['index', 'hash']);
            }
        }
        if (!this.isValidAddress(createBtcDelegationMessage.stakerAddr)) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid CreateBtcDelegationMessage stakerAddr: ${createBtcDelegationMessage.stakerAddr}`);
        }
    }
    validateWithdrawRewardMessage(withdrawRewardMessage) {
        if (withdrawRewardMessage._kind !== 'WithdrawReward') {
            throw new sdk_core_1.InvalidTransactionError(`Invalid WithdrawRewardMessage kind: ${withdrawRewardMessage._kind}`);
        }
        this.isObjPropertyNull(withdrawRewardMessage, ['type', 'address']);
        if (!['finality_provider', 'btc_staker'].includes(withdrawRewardMessage.type)) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid WithdrawRewardMessage type: ${withdrawRewardMessage.type}`);
        }
        if (!this.isValidAddress(withdrawRewardMessage.address)) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid WithdrawRewardMessage address: ${withdrawRewardMessage.address}`);
        }
    }
    /** @inheritdoc */
    isValidAddress(address) {
        return this.isValidCosmosLikeAddressWithMemoId(address, constants.accountAddressRegex);
    }
    /** @inheritdoc */
    isValidValidatorAddress(address) {
        return this.isValidBech32AddressMatchingRegex(address, constants.validatorAddressRegex);
    }
    /** @inheritdoc */
    isValidContractAddress(address) {
        return this.isValidBech32AddressMatchingRegex(address, constants.contractAddressRegex);
    }
    /** @inheritdoc */
    validateAmount(amount) {
        const amountBig = (0, bignumber_js_1.default)(amount.amount);
        if (amountBig.isLessThanOrEqualTo(0)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid amount: ' + amount.amount);
        }
        if (!constants.validDenoms.find((denom) => denom === amount.denom)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid denom: ' + amount.denom);
        }
    }
}
exports.Utils = Utils;
const utils = new Utils();
exports.default = utils;
