// @bitgo/sdk-coin-rune/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ROOT_PATH = exports.RUNE_FEES = exports.GAS_AMOUNT = exports.GAS_LIMIT = exports.TESTNET_ADDRESS_PREFIX = exports.testnetValidatorAddressRegex = exports.testnetAccountAddressRegex = exports.MAINNET_ADDRESS_PREFIX = exports.mainnetValidatorAddressRegex = exports.mainnetAccountAddressRegex = exports.validDenoms = void 0;
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const cosmosUtils = new abstract_cosmos_1.CosmosUtils();
exports.validDenoms = ['rune', ...cosmosUtils.getTokenDenomsUsingCoinFamily('thor')];
exports.mainnetAccountAddressRegex = /^(thor)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.mainnetValidatorAddressRegex = /^(thor)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.MAINNET_ADDRESS_PREFIX = 'thor';
exports.testnetAccountAddressRegex = /^(sthor)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.testnetValidatorAddressRegex = /^(sthor)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38})$/;
exports.TESTNET_ADDRESS_PREFIX = 'sthor';
exports.GAS_LIMIT = 200000;
exports.GAS_AMOUNT = '0'; // Gas amount should be zero for RUNE transactions, as fees (0.02 RUNE) is cut from sender balance directly in the transaction
exports.RUNE_FEES = '2000000'; // https://dev.thorchain.org/concepts/fees.html#thorchain-native-rune
exports.ROOT_PATH = 'm/0';
