// @bitgo/sdk-coin-rune/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuneUtils = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const constants = __importStar(require("./constants"));
const statics_1 = require("@bitgo/statics");
const constants_1 = require("./constants");
const bech32 = require('bech32-buffer');
class RuneUtils extends abstract_cosmos_1.CosmosUtils {
    constructor(networkType = statics_1.NetworkType.MAINNET) {
        super();
        this.networkType = networkType;
    }
    getSendMessageDataFromDecodedTx(decodedTx) {
        return decodedTx.body.messages.map((message) => {
            const value = this.registry.decode(message);
            return {
                value: {
                    fromAddress: this.getEncodedAddress(value.fromAddress),
                    toAddress: this.getEncodedAddress(value.toAddress),
                    amount: value.amount,
                },
                typeUrl: message.typeUrl,
            };
        });
    }
    /** @inheritdoc */
    isValidAddress(address) {
        if (address === undefined || address === null) {
            return false;
        }
        if (address instanceof Uint8Array) {
            return this.isValidDecodedAddress(address);
        }
        if (typeof address === 'string') {
            return this.isValidEncodedAddress(address);
        }
        return false;
    }
    /**
     * Validates a decoded address in `Uint8Array` form by encoding it and
     * checking if the encoded version is valid
     *
     * @param address - The decoded address as a `Uint8Array`.
     * @returns `true` if the encoded address is valid, `false` otherwise.
     */
    isValidDecodedAddress(address) {
        const encodedAddress = this.getEncodedAddress(address);
        return this.isValidEncodedAddress(encodedAddress);
    }
    /**
     * Validates an encoded address string against network-specific criteria.
     *
     * @param address - The encoded address as a `string`.
     * @returns `true` if the address meets network-specific validation criteria, `false` otherwise.
     */
    isValidEncodedAddress(address) {
        if (this.networkType === statics_1.NetworkType.TESTNET) {
            return this.isValidCosmosLikeAddressWithMemoId(address, constants.testnetAccountAddressRegex);
        }
        return this.isValidCosmosLikeAddressWithMemoId(address, constants.mainnetAccountAddressRegex);
    }
    /**
     * Encodes a given address `Uint8Array` into a bech32 string format, based on the current network type.
     * Primarily serves as a utility to convert a `Uint8Array`-type address to a bech32 encoded string
     *
     * @param address - The address to be encoded, provided as a `Uint8Array`.
     * @returns A bech32-encoded string representing the address.
     * @throws Error - Throws an error if encoding fails
     */
    getEncodedAddress(address) {
        try {
            return this.networkType === statics_1.NetworkType.TESTNET
                ? bech32.encode(constants_1.TESTNET_ADDRESS_PREFIX, address)
                : bech32.encode(constants_1.MAINNET_ADDRESS_PREFIX, address);
        }
        catch (error) {
            throw new Error(`Failed to encode address: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /**
     * Decodes a bech32-encoded address string back into a `Uint8Array`.
     * Primarily serves as a utility to convert a string-type address into its binary representation,
     *
     * @param address - The bech32-encoded address as a `string`.
     * @returns The decoded address as a `Uint8Array`.
     * @throws Error - Throws an error if decoding fails
     */
    getDecodedAddress(address) {
        try {
            return bech32.decode(address).data;
        }
        catch (error) {
            throw new Error(`Failed to decode address: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /** @inheritdoc */
    isValidValidatorAddress(address) {
        if (this.networkType === statics_1.NetworkType.TESTNET) {
            return this.isValidBech32AddressMatchingRegex(address, constants.testnetValidatorAddressRegex);
        }
        return this.isValidBech32AddressMatchingRegex(address, constants.mainnetValidatorAddressRegex);
    }
    /** @inheritdoc */
    validateAmount(amount) {
        const amountBig = (0, bignumber_js_1.default)(amount.amount);
        if (amountBig.isNaN() || amountBig.isLessThanOrEqualTo(0)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid amount: ' + amount.amount);
        }
        this.validateDenomination(amount.denom);
    }
    /**
     * Validates the gas limit and gas amount for a transaction.
     * @param {FeeData} gasBudget - The gas budget to validate.
     * @throws {InvalidTransactionError} Throws an error if the gas budget is invalid.
     */
    validateGasBudget(gasBudget) {
        if (gasBudget.gasLimit <= 0) {
            throw new sdk_core_1.InvalidTransactionError('Invalid gas limit ' + gasBudget.gasLimit);
        }
        this.validateGasAmountData(gasBudget.amount);
    }
    /**
     * Validates an array of coin amounts.
     * @param {Coin[]} amountArray - The array of coin amounts to validate.
     */
    validateGasAmountData(amountArray) {
        amountArray.forEach((coinAmount) => {
            this.validateGasAmount(coinAmount);
        });
    }
    validateGasAmount(amount) {
        const amountBig = (0, bignumber_js_1.default)(amount.amount);
        if (amountBig.isNaN() || amountBig.isLessThan(0)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid amount: ' + amount.amount);
        }
        this.validateDenomination(amount.denom);
    }
    validateDenomination(amountDenom) {
        if (!constants.validDenoms.find((denom) => denom === amountDenom)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid denom: ' + amountDenom);
        }
    }
    convertMessageAddressToUint8Array(messages) {
        return messages.map((message) => {
            if ('fromAddress' in message.value && 'toAddress' in message.value) {
                const sendMessage = message.value;
                const decodedFrom = typeof sendMessage.fromAddress === 'string'
                    ? bech32.decode(sendMessage.fromAddress).data
                    : sendMessage.fromAddress;
                const decodedTo = typeof sendMessage.toAddress === 'string' ? bech32.decode(sendMessage.toAddress).data : sendMessage.toAddress;
                return {
                    ...message,
                    value: {
                        ...sendMessage,
                        fromAddress: decodedFrom,
                        toAddress: decodedTo,
                    },
                };
            }
            return message;
        });
    }
    createTransaction(sequence, messages, gasBudget, publicKey, memo) {
        messages = this.convertMessageAddressToUint8Array(messages);
        const cosmosLikeTxn = {
            sequence: sequence,
            sendMessages: messages,
            gasBudget: gasBudget,
            publicKey: publicKey,
            memo: memo,
        };
        this.validateTransaction(cosmosLikeTxn);
        return cosmosLikeTxn;
    }
    getNetworkPrefix() {
        return this.networkType === statics_1.NetworkType.TESTNET ? constants_1.TESTNET_ADDRESS_PREFIX : constants_1.MAINNET_ADDRESS_PREFIX;
    }
}
exports.RuneUtils = RuneUtils;
const runeUtils = new RuneUtils();
exports.default = runeUtils;
