// @bitgo/abstract-utxo/dist/esm/recovery/baseApi.js
import superagent from 'superagent';
import { BitGoJsError } from '@bitgo/sdk-core';
export class ApiNotImplementedError extends BitGoJsError {
    constructor(coinName) {
        super(`api not implemented for coin ${coinName}`);
    }
}
export class ApiRequestError extends BitGoJsError {
    constructor(url, reason) {
        super(`error in API request: ${url}: ${reason}`);
    }
    static forRequest(req, reason) {
        return new ApiRequestError(req.toString(), reason);
    }
    static forResponse(res, reason) {
        return new ApiRequestError(res.header, reason);
    }
}
export class BaseApi {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
        if (!baseUrl.startsWith('https://')) {
            throw new Error(`baseUrl must start with https://`);
        }
        if (baseUrl.endsWith(`/`)) {
            throw new Error(`baseUrl must not end with slash (/)`);
        }
    }
    async request(method, path, requestBody, params = {}) {
        if (!path.startsWith(path)) {
            throw new Error(`path must start with '/' character`);
        }
        const url = this.baseUrl + path;
        let response;
        try {
            response = await superagent(method, url)
                .retry(params.retry)
                .send(requestBody);
        }
        catch (e) {
            throw new ApiRequestError(url, e);
        }
        if (!response.ok) {
            throw new ApiRequestError(url, `status code ${response.statusCode}`);
        }
        return {
            map(f) {
                try {
                    return f(response.body);
                }
                catch (e) {
                    throw new ApiRequestError(url, e);
                }
            },
        };
    }
    get(path, params) {
        return this.request('get', path, undefined, params);
    }
    post(path, body, params) {
        return this.request('post', path, body, params);
    }
}
