// @bitgo/abstract-utxo/dist/esm/recovery/crossChainRecovery.js
import { BIP32, fixedScriptWallet, address as wasmAddress } from '@bitgo/wasm-utxo';
import { Wallet } from '@bitgo/sdk-core';
import { decrypt } from '@bitgo/sdk-api';
import { signAndVerifyPsbt } from '../transaction/fixedScript/signTransaction';
import { encodeTransaction } from '../transaction/decode';
import { getReplayProtectionPubkeys } from '../transaction/fixedScript/replayProtection';
import { toTNumber } from '../tnumber';
import { unspentSum } from '../unspent';
import { createEmptyWasmPsbt, addWalletInputsToWasmPsbt, addOutputToWasmPsbt, getRecoveryAmount } from './psbt';
export async function getWallet(bitgo, coin, walletId) {
    try {
        return await coin.wallets().get({ id: walletId });
    }
    catch (e) {
        // TODO: BG-46364 handle errors more gracefully
        // The v2 endpoint coin.wallets().get() may throw 404 or 400 errors, but this should not prevent us from searching for the walletId in v1 wallets.
        if (e.status >= 500) {
            throw e;
        }
    }
    try {
        return await bitgo.wallets().get({ id: walletId });
    }
    catch (e) {
        throw new Error(`could not get wallet ${walletId} from v1 or v2: ${e.toString()}`);
    }
}
/**
 * @param recoveryCoin
 * @param wallet
 * @return wallet pubkeys
 */
export async function getWalletKeys(recoveryCoin, wallet) {
    let xpubs;
    if (wallet instanceof Wallet) {
        const keychains = (await recoveryCoin.keychains().getKeysForSigning({ wallet }));
        if (keychains.length !== 3) {
            throw new Error(`expected triple got ${keychains.length}`);
        }
        xpubs = keychains.map((k) => k.pub);
    }
    else {
        xpubs = wallet.keychains.map((k) => k.xpub);
    }
    return fixedScriptWallet.RootWalletKeys.from(xpubs);
}
export async function isWalletAddress(wallet, address) {
    try {
        let addressData;
        if (wallet instanceof Wallet) {
            addressData = await wallet.getAddress({ address });
        }
        else {
            addressData = await wallet.address({ address });
        }
        return addressData !== undefined;
    }
    catch (e) {
        return false;
    }
}
/**
 * Convert a Litecoin P2SH address from M... format (scriptHash 0x32) to the legacy 3... format (scriptHash 0x05)
 * and Q... format (scriptHash 0x32) to the legacy 2... format (scriptHash 0x05).
 * This is needed for cross-chain recovery when (T)LTC was sent to a (T)BTC address, because the (T)BTC wallet
 * stores addresses in the 3... format while the LTC blockchain returns addresses in M... format.
 *
 * @param address - LTC address to convert
 * @param coinName - The coin name (e.g. 'ltc', 'tltc')
 * @returns The address in legacy 3... format, or the original address if it's not a P2SH address
 */
export function convertLtcAddressToLegacyFormat(address, coinName) {
    if (coinName !== 'ltc' && coinName !== 'tltc') {
        return address;
    }
    // only consider addresses that start with M or Q (modern ltc p2sh prefixes)
    if (!address.startsWith('M') && !address.startsWith('Q')) {
        return address;
    }
    const scriptPubKey = wasmAddress.toOutputScriptWithCoin(address, coinName);
    const dstCoin = coinName === 'ltc' ? 'btc' : 'tbtc';
    const dstAddress = wasmAddress.fromOutputScriptWithCoin(scriptPubKey, dstCoin);
    return dstAddress;
}
/**
 * @param coin
 * @param txid
 * @param amountType
 * @param wallet
 * @param apiKey - a blockchair api key
 * @return all unspents for transaction outputs, including outputs from other transactions
 */
async function getAllRecoveryOutputs(coin, txid, amountType = 'number', wallet, apiKey) {
    const api = coin.getRecoveryProvider(apiKey);
    const tx = await api.getTransactionIO(txid);
    const walletAddresses = (await Promise.all(tx.outputs.map(async (output) => {
        // For some coins (bch) we need to convert the address to legacy format since the api returns the address
        // in non legacy format. However, we want to keep the address in the same format as the response since we
        // are going to hit the API again to fetch address unspents.
        const canonicalAddress = coin.canonicalAddress(output.address);
        let isWalletOwned = await isWalletAddress(wallet, canonicalAddress);
        // For LTC cross-chain recovery: if the address isn't found, try the legacy format.
        // When LTC is sent to a BTC address, the LTC blockchain returns M... addresses
        // but the BTC wallet stores addresses in 3... format.
        if (!isWalletOwned && coin.getFamily() === 'ltc') {
            const legacyAddress = convertLtcAddressToLegacyFormat(output.address, coin.name);
            if (legacyAddress !== output.address) {
                isWalletOwned = await isWalletAddress(wallet, legacyAddress);
            }
        }
        return isWalletOwned ? output.address : null;
    }))).filter((address) => address !== null);
    const unspents = await api.getUnspentsForAddresses(walletAddresses);
    if (unspents.length === 0) {
        throw new Error(`No recovery unspents found.`);
    }
    // the api may return cashaddr's instead of legacy for BCH and BCHA
    // downstream processes's only expect legacy addresses
    return unspents.map((recoveryOutput) => {
        return {
            ...recoveryOutput,
            address: coin.canonicalAddress(recoveryOutput.address),
            value: toTNumber(BigInt(recoveryOutput.value), amountType),
        };
    });
}
async function getScriptId(coin, wallet, script) {
    const address = wasmAddress.fromOutputScriptWithCoin(script, coin.name);
    let addressData;
    if (wallet instanceof Wallet) {
        addressData = await wallet.getAddress({ address });
    }
    else {
        addressData = await wallet.address({ address });
    }
    if (typeof addressData.chain === 'number' && typeof addressData.index === 'number') {
        return { chain: addressData.chain, index: addressData.index };
    }
    throw new Error(`invalid address data: ${JSON.stringify(addressData)}`);
}
/**
 * Lookup address data from unspents on sourceCoin in address database of recoveryCoin.
 * Return full walletUnspents including scriptId in sourceCoin format.
 *
 * @param sourceCoin
 * @param recoveryCoin
 * @param unspents
 * @param wallet
 * @return walletUnspents
 */
async function toWalletUnspents(sourceCoin, recoveryCoin, unspents, wallet) {
    const addresses = new Set(unspents.map((u) => u.address));
    const walletUnspents = [];
    for (const address of addresses) {
        let scriptId;
        try {
            scriptId = await getScriptId(recoveryCoin, wallet, wasmAddress.toOutputScriptWithCoin(address, sourceCoin.name));
        }
        catch (e) {
            console.error(`error getting scriptId for ${address}:`, e);
            continue;
        }
        const filteredUnspents = unspents
            .filter((u) => u.address === address)
            .map((u) => ({
            ...u,
            ...scriptId,
        }));
        walletUnspents.push(...filteredUnspents);
    }
    return walletUnspents;
}
/**
 * @param coin
 * @return feeRate for transaction
 */
async function getFeeRateSatVB(coin) {
    // TODO: use feeRate API
    const feeRate = {
        bch: 20,
        tbch: 20,
        bcha: 20,
        tbcha: 20,
        bsv: 20,
        tbsv: 20,
        btc: 80,
        tbtc: 80,
        tbtcsig: 80,
        tbtc4: 80,
        tbtcbgsig: 80,
        ltc: 100,
        tltc: 100,
        doge: 1000,
        tdoge: 1000,
    }[coin.getChain()];
    if (!feeRate) {
        throw new Error(`no feeRate for ${coin.getChain()}`);
    }
    return feeRate;
}
/**
 * @param xprv
 * @param passphrase
 * @param wallet
 * @return signing key
 */
async function getPrv(xprv, passphrase, wallet) {
    if (xprv) {
        const key = BIP32.fromBase58(xprv);
        if (key.isNeutered()) {
            throw new Error(`not a private key`);
        }
        return key;
    }
    if (!wallet || !passphrase) {
        throw new Error(`no xprv given: need wallet and passphrase to continue`);
    }
    let encryptedPrv;
    if (wallet instanceof Wallet) {
        encryptedPrv = (await wallet.getEncryptedUserKeychain()).encryptedPrv;
    }
    else {
        encryptedPrv = (await wallet.getEncryptedUserKeychain()).encryptedXprv;
    }
    return getPrv(decrypt(passphrase, encryptedPrv));
}
/**
 * Create a sweep transaction for cross-chain recovery using wasm-utxo
 * @param coinName - BitGo coin name (e.g. 'btc', 'tbtc', 'ltc')
 * @param walletKeys
 * @param unspents
 * @param targetAddress
 * @param feeRateSatVB
 * @return unsigned PSBT
 */
function createSweepTransaction(coinName, walletKeys, unspents, targetAddress, feeRateSatVB) {
    const inputValue = unspentSum(unspents);
    // Create PSBT with wasm-utxo and add wallet inputs using shared utilities
    const unspentsBigint = unspents.map((u) => ({ ...u, value: BigInt(u.value) }));
    const wasmPsbt = createEmptyWasmPsbt(coinName, walletKeys);
    addWalletInputsToWasmPsbt(wasmPsbt, unspentsBigint, walletKeys);
    // Calculate dimensions using wasm-utxo Dimensions
    const vsize = fixedScriptWallet.Dimensions.fromPsbt(wasmPsbt)
        .plus(fixedScriptWallet.Dimensions.fromOutput(targetAddress, coinName))
        .getVSize();
    const fee = BigInt(Math.round(vsize * feeRateSatVB));
    // Add output to wasm PSBT
    addOutputToWasmPsbt(wasmPsbt, targetAddress, inputValue - fee, coinName);
    return wasmPsbt;
}
/**
 * Recover wallet deposits that were received on the wrong blockchain
 * (for instance bitcoin deposits that were received for a litecoin wallet).
 *
 * Fetches the unspent data from BitGo's public blockchain API and the script data from the user's
 * wallet.
 *
 * @param {BitGoBase} bitgo
 * @param {RecoverParams} params
 */
export async function recoverCrossChain(bitgo, params) {
    const wallet = await getWallet(bitgo, params.recoveryCoin, params.walletId);
    const unspents = await getAllRecoveryOutputs(params.sourceCoin, params.txid, params.sourceCoin.amountType, wallet, params.apiKey);
    const walletUnspents = await toWalletUnspents(params.sourceCoin, params.recoveryCoin, unspents, wallet);
    const walletKeys = fixedScriptWallet.RootWalletKeys.from(await getWalletKeys(params.recoveryCoin, wallet));
    const prv = params.xprv || params.walletPassphrase ? await getPrv(params.xprv, params.walletPassphrase, wallet) : undefined;
    const feeRateSatVB = await getFeeRateSatVB(params.sourceCoin);
    // Create PSBT for both signed and unsigned recovery
    let psbt = createSweepTransaction(params.sourceCoin.getChain(), walletKeys, walletUnspents, params.recoveryAddress, feeRateSatVB);
    // For unsigned recovery, return unsigned PSBT hex
    if (!prv) {
        return {
            txHex: encodeTransaction(psbt).toString('hex'),
            walletId: params.walletId,
            address: params.recoveryAddress,
            coin: params.sourceCoin.getChain(),
        };
    }
    // For signed recovery, sign the PSBT with user key and return half-signed PSBT
    psbt = signAndVerifyPsbt(psbt, prv, walletKeys, {
        publicKeys: getReplayProtectionPubkeys(params.sourceCoin.name),
    });
    return {
        version: wallet instanceof Wallet ? 2 : 1,
        walletId: params.walletId,
        txHex: encodeTransaction(psbt).toString('hex'),
        sourceCoin: params.sourceCoin.getChain(),
        recoveryCoin: params.recoveryCoin.getChain(),
        recoveryAmount: toTNumber(getRecoveryAmount(psbt, walletKeys, params.recoveryAddress), params.sourceCoin.amountType),
    };
}
