// @iota/iota-sdk/dist/cjs/client/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var client_exports = {};
__export(client_exports, {
  IotaClient: () => import_client.IotaClient,
  IotaHTTPStatusError: () => import_errors.IotaHTTPStatusError,
  IotaHTTPTransport: () => import_http_transport.IotaHTTPTransport,
  IotaHTTPTransportError: () => import_errors.IotaHTTPTransportError,
  JsonRpcError: () => import_errors.JsonRpcError,
  Network: () => import_network.Network,
  getAllNetworks: () => import_network.getAllNetworks,
  getDefaultNetwork: () => import_network.getDefaultNetwork,
  getFullnodeUrl: () => import_network.getFullnodeUrl,
  getGraphQLUrl: () => import_network.getGraphQLUrl,
  getNetwork: () => import_network.getNetwork,
  isIotaClient: () => import_client.isIotaClient
});
module.exports = __toCommonJS(client_exports);
var import_http_transport = require("./http-transport.js");
var import_network = require("./network.js");
var import_client = require("./client.js");
var import_errors = require("./errors.js");
__reExport(client_exports, require("./apps-backend.js"), module.exports);
