// @bitgo/sdk-coin-tao/dist/src/tao.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tao = exports.DEFAULT_SCAN_FACTOR = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const lib_1 = require("./lib");
const api_1 = require("@polkadot/api");
exports.DEFAULT_SCAN_FACTOR = 20; // default number of receive addresses to scan for funds
class Tao extends abstract_substrate_1.SubstrateCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this.staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Tao(bitgo, staticsCoin);
    }
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    getMaxValidityDurationBlocks() {
        return 2400;
    }
    allowsAccountConsolidations() {
        return true;
    }
    async getInitializedNodeAPI() {
        if (!Tao.nodeApiInitialized) {
            const wsProvider = new api_1.WsProvider(sdk_core_1.Environments[this.bitgo.getEnv()].substrateNodeUrls);
            Tao.API = await api_1.ApiPromise.create({ provider: wsProvider });
            Tao.nodeApiInitialized = true;
        }
        return Tao.API;
    }
    async getAccountInfo(walletAddr) {
        const api = await this.getInitializedNodeAPI();
        const { nonce, data: balance } = await api.query.system.account(walletAddr);
        return { nonce: nonce.toNumber(), freeBalance: balance.free.toNumber() };
    }
    async getFee(destAddr, srcAddr, amount) {
        const api = await this.getInitializedNodeAPI();
        const info = await api.tx.balances.transferAllowDeath(destAddr, amount).paymentInfo(srcAddr);
        return info.partialFee.toNumber();
    }
    async getHeaderInfo() {
        const api = await this.getInitializedNodeAPI();
        const { number, hash } = await api.rpc.chain.getHeader();
        return { headerNumber: number.toNumber(), headerHash: hash.toString() };
    }
    async getMaterial() {
        const api = await this.getInitializedNodeAPI();
        return {
            genesisHash: api.genesisHash.toString(),
            chainName: api.runtimeChain.toString(),
            specName: api.runtimeVersion.specName.toString(),
            specVersion: api.runtimeVersion.specVersion.toNumber(),
            txVersion: api.runtimeVersion.transactionVersion.toNumber(),
            metadata: api.runtimeMetadata.toHex(),
        };
    }
}
exports.Tao = Tao;
Tao.nodeApiInitialized = false;
