// @bitgo/sdk-coin-tao/dist/src/taoToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaoToken = void 0;
const statics_1 = require("@bitgo/statics");
const tao_1 = require("./tao");
class TaoToken extends tao_1.Tao {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('tao') : statics_1.coins.get('ttao');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new TaoToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.tao.tokens, ...statics_1.tokens.testnet.tao.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            const tokenConstructor = TaoToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get subnetId() {
        return this.tokenConfig.subnetId;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return this._staticsCoin.fullName;
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.TaoToken = TaoToken;
