// @bitgo/logger/dist/test/unit/sanitizeLog.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const sanitizeLog_1 = require("../../src/sanitizeLog");
const V2_TOKEN = 'v2xaabbccdd112233445566778899aabbccddeeff';
const XPRV_KEY = 'xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi';
describe('sanitize', function () {
    describe('primitives', function () {
        it('should pass through null', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(null), null);
        });
        it('should pass through undefined', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(undefined), undefined);
        });
        it('should pass through numbers', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(42), 42);
        });
        it('should pass through booleans', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(true), true);
        });
        it('should pass through plain strings', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)('hello world'), 'hello world');
        });
    });
    describe('sensitive string value redaction', function () {
        it('should redact a string starting with v2x', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(V2_TOKEN), '<REMOVED>');
        });
        it('should redact a short string starting with v2x', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)('v2xaabb'), '<REMOVED>');
        });
        it('should redact a string starting with xprv', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(XPRV_KEY), '<REMOVED>');
        });
        it('should not redact when sensitive prefix is not at the start', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(`Bearer ${V2_TOKEN}`), `Bearer ${V2_TOKEN}`);
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)('not an xprv key'), 'not an xprv key');
        });
        it('should not redact a plain string without sensitive prefix', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)('normal message'), 'normal message');
        });
    });
    describe('sensitive key redaction', function () {
        it('should redact token key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ token: 'abc' }), { token: '<REMOVED>' });
        });
        it('should redact password key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ password: 'secret' }), { password: '<REMOVED>' });
        });
        it('should redact prv key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ prv: 'xprv123' }), { prv: '<REMOVED>' });
        });
        it('should redact xprv key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ xprv: 'xprvkey' }), { xprv: '<REMOVED>' });
        });
        it('should redact privateKey key (case-insensitive)', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ privateKey: 'key' }), { privateKey: '<REMOVED>' });
        });
        it('should redact otp key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ otp: '123456' }), { otp: '<REMOVED>' });
        });
        it('should redact passphrase key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ passphrase: 'secret' }), { passphrase: '<REMOVED>' });
        });
        it('should redact walletPassphrase key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ walletPassphrase: 'pass' }), { walletPassphrase: '<REMOVED>' });
        });
        it('should redact bearer key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ bearer: 'token123' }), { bearer: '<REMOVED>' });
        });
        it('should redact _token key', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ _token: 'abc' }), { _token: '<REMOVED>' });
        });
        it('should keep non-sensitive keys', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ user: 'alice' }), { user: 'alice' });
        });
    });
    describe('sensitive string value redaction in object values', function () {
        it('should redact an object value starting with v2x', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ auth: V2_TOKEN }), { auth: '<REMOVED>' });
        });
        it('should redact an object value starting with xprv', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ key: XPRV_KEY }), { key: '<REMOVED>' });
        });
        it('should redact a short v2x object value', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ key: 'v2xaabb' }), { key: '<REMOVED>' });
        });
        it('should not redact when sensitive prefix is not at the start of value', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ msg: `token is ${V2_TOKEN}` }), { msg: `token is ${V2_TOKEN}` });
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ header: `Bearer ${V2_TOKEN}` }), { header: `Bearer ${V2_TOKEN}` });
        });
        it('should keep an object value without sensitive prefix', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ msg: 'hello' }), { msg: 'hello' });
        });
    });
    describe('sensitive string values in arrays', function () {
        it('should redact a v2x string in an array', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)([V2_TOKEN, 'hello']), ['<REMOVED>', 'hello']);
        });
        it('should redact an xprv string in an array', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)([XPRV_KEY, 'hello']), ['<REMOVED>', 'hello']);
        });
    });
    describe('nested objects with sensitive keys', function () {
        it('should redact sensitive keys at any depth', function () {
            const obj = {
                user: 'alice',
                credentials: {
                    password: 'secret',
                    otp: '123456',
                    config: {
                        prv: 'private-key',
                        endpoint: 'https://api.bitgo.com',
                    },
                },
            };
            const result = (0, sanitizeLog_1.sanitize)(obj);
            assert_1.default.strictEqual(result.user, 'alice');
            assert_1.default.strictEqual(result.credentials.password, '<REMOVED>');
            assert_1.default.strictEqual(result.credentials.otp, '<REMOVED>');
            assert_1.default.strictEqual(result.credentials.config.prv, '<REMOVED>');
            assert_1.default.strictEqual(result.credentials.config.endpoint, 'https://api.bitgo.com');
        });
    });
    describe('nested Error inside objects', function () {
        it('should preserve Error properties nested in an object', function () {
            const obj = { action: 'transfer', error: new Error('nested failure') };
            const result = (0, sanitizeLog_1.sanitize)(obj);
            assert_1.default.strictEqual(result.error.name, 'Error');
            assert_1.default.strictEqual(result.error.message, 'nested failure');
            assert_1.default.strictEqual(typeof result.error.stack, 'string');
        });
        it('should preserve Error properties at any depth', function () {
            const obj = { data: { inner: { error: new Error('deep') } } };
            const result = (0, sanitizeLog_1.sanitize)(obj);
            assert_1.default.strictEqual(result.data.inner.error.message, 'deep');
        });
        it('should redact sensitive custom properties on nested Errors', function () {
            const err = new Error('fail');
            err.token = 'secret';
            err.statusCode = 500;
            const obj = { error: err };
            const result = (0, sanitizeLog_1.sanitize)(obj);
            assert_1.default.strictEqual(result.error.token, '<REMOVED>');
            assert_1.default.strictEqual(result.error.statusCode, 500);
            assert_1.default.strictEqual(result.error.message, 'fail');
        });
    });
    describe('Date handling', function () {
        it('should convert a top-level Date to ISO string', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(new Date('2026-03-02T10:30:00.000Z')), '2026-03-02T10:30:00.000Z');
        });
        it('should convert a Date inside an object to ISO string', function () {
            const obj = { createdAt: new Date('2026-01-15T00:00:00.000Z'), user: 'bob' };
            const result = (0, sanitizeLog_1.sanitize)(obj);
            assert_1.default.strictEqual(result.createdAt, '2026-01-15T00:00:00.000Z');
            assert_1.default.strictEqual(result.user, 'bob');
        });
        it('should handle an invalid Date without throwing', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(new Date('not-a-date')), '[Invalid Date]');
        });
    });
    describe('BigInt handling', function () {
        it('should convert a top-level BigInt to string', function () {
            assert_1.default.strictEqual((0, sanitizeLog_1.sanitize)(100n), '100');
        });
        it('should convert BigInts inside an object to strings', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)({ amount: 100n, fee: 50n }), { amount: '100', fee: '50' });
        });
        it('should convert BigInts inside an array to strings', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)([1n, 2n, 3n]), ['1', '2', '3']);
        });
    });
    describe('circular reference handling', function () {
        it('should replace circular references with [Circular]', function () {
            const circular = { name: 'test' };
            circular.self = circular;
            const result = (0, sanitizeLog_1.sanitize)(circular);
            assert_1.default.strictEqual(result.name, 'test');
            assert_1.default.strictEqual(result.self, '[Circular]');
        });
    });
    describe('array handling', function () {
        it('should pass through a simple array', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)([1, 'hello', true]), [1, 'hello', true]);
        });
        it('should sanitize objects inside arrays', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)([{ password: 'secret' }, { user: 'alice' }]), [
                { password: '<REMOVED>' },
                { user: 'alice' },
            ]);
        });
        it('should handle nested arrays', function () {
            assert_1.default.deepStrictEqual((0, sanitizeLog_1.sanitize)([
                [1, 2],
                [3, 4],
            ]), [
                [1, 2],
                [3, 4],
            ]);
        });
    });
    describe('max depth handling', function () {
        it('should return [Max Depth Exceeded] beyond depth limit', function () {
            let deepObj = { value: 'bottom' };
            for (let i = 0; i < 30; i++) {
                deepObj = { nested: deepObj };
            }
            const result = (0, sanitizeLog_1.sanitize)(deepObj);
            let current = result;
            let reachedLimit = false;
            for (let i = 0; i < 30; i++) {
                if (current.nested === '[Max Depth Exceeded]') {
                    reachedLimit = true;
                    break;
                }
                current = current.nested;
            }
            assert_1.default.strictEqual(reachedLimit, true);
        });
    });
    describe('mixed complex object', function () {
        it('should handle all data types together', function () {
            const complexObj = {
                user: 'alice',
                password: 'secret',
                transaction: {
                    amount: 100n,
                    createdAt: new Date('2026-06-15T12:00:00.000Z'),
                    error: new Error('validation failed'),
                    token: 'my-token',
                },
                authToken: V2_TOKEN,
                privateKey: XPRV_KEY,
                tags: ['transfer', 'urgent'],
            };
            const result = (0, sanitizeLog_1.sanitize)(complexObj);
            assert_1.default.strictEqual(result.password, '<REMOVED>');
            assert_1.default.strictEqual(result.transaction.amount, '100');
            assert_1.default.strictEqual(result.transaction.createdAt, '2026-06-15T12:00:00.000Z');
            assert_1.default.strictEqual(result.transaction.error.message, 'validation failed');
            assert_1.default.strictEqual(result.transaction.token, '<REMOVED>');
            assert_1.default.strictEqual(result.authToken, '<REMOVED>');
            assert_1.default.strictEqual(result.privateKey, '<REMOVED>');
            assert_1.default.strictEqual(result.tags[0], 'transfer');
            assert_1.default.strictEqual(result.tags[1], 'urgent');
        });
    });
});
describe('getErrorData', function () {
    it('should extract name, message, and stack from an Error', function () {
        const err = new Error('something broke');
        const result = (0, sanitizeLog_1.getErrorData)(err);
        assert_1.default.strictEqual(result.name, 'Error');
        assert_1.default.strictEqual(result.message, 'something broke');
        assert_1.default.strictEqual(typeof result.stack, 'string');
    });
    it('should preserve custom enumerable properties', function () {
        const err = new Error('auth failed');
        err.statusCode = 401;
        err.url = 'https://api.bitgo.com';
        const result = (0, sanitizeLog_1.getErrorData)(err);
        assert_1.default.strictEqual(result.statusCode, 401);
        assert_1.default.strictEqual(result.url, 'https://api.bitgo.com');
        assert_1.default.strictEqual(result.name, 'Error');
        assert_1.default.strictEqual(result.message, 'auth failed');
    });
    it('should return non-Error values as-is', function () {
        assert_1.default.strictEqual((0, sanitizeLog_1.getErrorData)('hello'), 'hello');
        assert_1.default.strictEqual((0, sanitizeLog_1.getErrorData)(null), null);
        assert_1.default.strictEqual((0, sanitizeLog_1.getErrorData)(42), 42);
        assert_1.default.strictEqual((0, sanitizeLog_1.getErrorData)(undefined), undefined);
    });
    it('should extract Error.cause when present', function () {
        const inner = new Error('db connection failed');
        const outer = new Error('transaction failed', { cause: inner });
        const result = (0, sanitizeLog_1.getErrorData)(outer);
        assert_1.default.strictEqual(result.message, 'transaction failed');
        const causeData = result.cause;
        assert_1.default.strictEqual(causeData.name, 'Error');
        assert_1.default.strictEqual(causeData.message, 'db connection failed');
        assert_1.default.strictEqual(typeof causeData.stack, 'string');
    });
    it('should extract non-enumerable custom properties', function () {
        const err = new Error('fail');
        Object.defineProperty(err, 'code', { value: 'ECONNREFUSED', enumerable: false });
        const result = (0, sanitizeLog_1.getErrorData)(err);
        assert_1.default.strictEqual(result.code, 'ECONNREFUSED');
        assert_1.default.strictEqual(result.message, 'fail');
    });
    it('should handle nested Error causes', function () {
        const root = new Error('root cause');
        const middle = new Error('middle', { cause: root });
        const outer = new Error('outer', { cause: middle });
        const result = (0, sanitizeLog_1.getErrorData)(outer);
        const middleData = result.cause;
        assert_1.default.strictEqual(middleData.message, 'middle');
        const rootData = middleData.cause;
        assert_1.default.strictEqual(rootData.message, 'root cause');
    });
    describe('full Error sanitization flow (getErrorData + sanitize)', function () {
        it('should preserve Error info and redact sensitive custom properties', function () {
            const err = new Error('failed');
            err.token = 'secret-token';
            err.password = 'hunter2';
            err.statusCode = 500;
            const result = (0, sanitizeLog_1.sanitize)((0, sanitizeLog_1.getErrorData)(err));
            assert_1.default.strictEqual(result.name, 'Error');
            assert_1.default.strictEqual(result.message, 'failed');
            assert_1.default.strictEqual(typeof result.stack, 'string');
            assert_1.default.strictEqual(result.token, '<REMOVED>');
            assert_1.default.strictEqual(result.password, '<REMOVED>');
            assert_1.default.strictEqual(result.statusCode, 500);
        });
        it('should redact sensitive string values in custom Error properties', function () {
            const err = new Error('bad request');
            err.authHeader = V2_TOKEN;
            err.key = XPRV_KEY;
            const result = (0, sanitizeLog_1.sanitize)((0, sanitizeLog_1.getErrorData)(err));
            assert_1.default.strictEqual(result.authHeader, '<REMOVED>');
            assert_1.default.strictEqual(result.key, '<REMOVED>');
            assert_1.default.strictEqual(result.message, 'bad request');
        });
    });
});
