// @bitgo/abstract-lightning/dist/src/codecs/api/balance.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LndGetBalancesResponse = exports.LndGetOnchainBalances = exports.LndGetOffchainBalances = exports.LndBalance = exports.LndWalletBalance = exports.ChannelBalance = exports.LndAmount = void 0;
const t = __importStar(require("io-ts"));
const BigIntFromString_1 = require("io-ts-types/BigIntFromString");
// codecs for lightning wallet balance api
exports.LndAmount = t.strict({
    sat: BigIntFromString_1.BigIntFromString,
    msat: BigIntFromString_1.BigIntFromString,
}, 'LndAmount');
exports.ChannelBalance = t.strict({
    /** The balance on your side of the channel and what you the user can spend. */
    localBalance: exports.LndAmount,
    /** The balance on the other side of the channel, what your channel partner can controls. */
    remoteBalance: exports.LndAmount,
    /** Sum of local unsettled balances. */
    unsettledLocalBalance: exports.LndAmount,
    /** Sum of remote unsettled balances. */
    unsettledRemoteBalance: exports.LndAmount,
    /** Sum of local pending balances. */
    pendingOpenLocalBalance: exports.LndAmount,
    /** Sum of local remote balances. */
    pendingOpenRemoteBalance: exports.LndAmount,
    /** Total local reserved balance. */
    reservedLocalBalance: exports.LndAmount,
    /** Total local spendable balance. */
    spendableLocalBalance: exports.LndAmount,
}, 'ChannelBalance');
exports.LndWalletBalance = t.strict({
    /** Total balance, confirmed and unconfirmed */
    totalBalance: BigIntFromString_1.BigIntFromString,
    confirmedBalance: BigIntFromString_1.BigIntFromString,
    unconfirmedBalance: BigIntFromString_1.BigIntFromString,
    lockedBalance: BigIntFromString_1.BigIntFromString,
    reservedBalanceAnchorChan: BigIntFromString_1.BigIntFromString,
}, 'LndWalletBalance');
/**
 The balances as returned from lnd.

 Wallet Balance
 https://api.lightning.community/api/lnd/lightning/wallet-balance/index.html

 Channel Balance
 https://api.lightning.community/api/lnd/lightning/channel-balance/index.html
 */
exports.LndBalance = t.strict({
    offchain: exports.ChannelBalance,
    onchain: exports.LndWalletBalance,
    totalLimboBalance: BigIntFromString_1.BigIntFromString,
}, 'LndBalance');
exports.LndGetOffchainBalances = t.strict({
    outboundBalance: BigIntFromString_1.BigIntFromString,
    outboundPendingBalance: BigIntFromString_1.BigIntFromString,
    outboundUnsettledBalance: BigIntFromString_1.BigIntFromString,
    outboundLockedBalance: BigIntFromString_1.BigIntFromString,
    outboundReservedBalance: BigIntFromString_1.BigIntFromString,
    outboundSpendableBalance: BigIntFromString_1.BigIntFromString,
    inboundBalance: BigIntFromString_1.BigIntFromString,
    inboundPendingBalance: BigIntFromString_1.BigIntFromString,
    inboundUnsettledBalance: BigIntFromString_1.BigIntFromString,
}, 'LndGetOffchainBalances');
// wallet onchain balances, names forced by type in AbstractCoin
exports.LndGetOnchainBalances = t.strict({
    spendableBalanceString: BigIntFromString_1.BigIntFromString,
    balanceString: BigIntFromString_1.BigIntFromString,
    confirmedBalanceString: BigIntFromString_1.BigIntFromString,
}, 'LndGetOnchainBalances');
exports.LndGetBalancesResponse = t.intersection([
    t.strict({
        offchain: exports.LndGetOffchainBalances,
    }),
    exports.LndGetOnchainBalances,
], 'LndGetBalancesResponse');
