// @bitgo/abstract-lightning/dist/src/codecs/api/invoice.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvoiceQuery = exports.ListInvoicesResponse = exports.Invoice = exports.CreateInvoiceBody = exports.InvoiceStatus = void 0;
const t = __importStar(require("io-ts"));
const BigIntFromString_1 = require("io-ts-types/BigIntFromString");
const DateFromISOString_1 = require("io-ts-types/DateFromISOString");
// codecs for lightning wallet invoice related apis
exports.InvoiceStatus = t.union([
    // Initial state.
    // Transitions to 'settled' or 'canceled' on LND notification.
    t.literal('open'),
    // Final state.
    t.literal('settled'),
    // Final state.
    t.literal('canceled'),
], 'InvoiceStatus');
exports.CreateInvoiceBody = t.partial({
    valueMsat: BigIntFromString_1.BigIntFromString,
    valueSat: t.number,
    memo: t.string,
    expiry: t.number,
});
/**
 * A representation of the data tracked by the indexer for an invoice it has
 * created within lnd.
 */
exports.Invoice = t.intersection([
    t.type({
        valueMsat: BigIntFromString_1.BigIntFromString,
        paymentHash: t.string,
        /** The BOLT #11 encoded invoice string */
        invoice: t.string,
        /** The public BitGo walletId to which this invoice belongs */
        walletId: t.string,
        status: exports.InvoiceStatus,
        /** A date in ISO format representing when this invoice expires. */
        expiresAt: DateFromISOString_1.DateFromISOString,
        createdAt: DateFromISOString_1.DateFromISOString,
        updatedAt: DateFromISOString_1.DateFromISOString,
    }),
    t.partial({
        // Keep valueSat for backward compatibility with older versions of the API
        // Will be moved up to required fields in the future
        valueSat: t.number,
        memo: t.string,
        amtPaidMsat: BigIntFromString_1.BigIntFromString,
    }),
], 'Invoice');
exports.ListInvoicesResponse = t.intersection([
    t.type({
        invoices: t.array(exports.Invoice),
    }),
    t.partial({
        /**
         * This is the paymentHash of the last Invoice in the last iteration.
         * Providing this value as the prevId in the next request will return the next batch of invoices.
         * */
        nextBatchPrevId: t.string,
    }),
], 'ListInvoicesResponse');
exports.InvoiceQuery = t.partial({
    status: exports.InvoiceStatus,
    limit: BigIntFromString_1.BigIntFromString,
    /** paymentHash provided by nextBatchPrevId in the previous list */
    prevId: t.string,
    startDate: DateFromISOString_1.DateFromISOString,
    endDate: DateFromISOString_1.DateFromISOString,
}, 'InvoiceQuery');
