// @bitgo/abstract-lightning/dist/src/codecs/api/payment.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LndCreatePaymentResponse = exports.SubmitPaymentParams = exports.PaymentQuery = exports.ListPaymentsResponse = exports.PaymentInfo = exports.PaymentFailureReason = exports.PAYMENT_FAILURE_FORCE_FAILED = exports.PAYMENT_FAILURE_CANCELED = exports.PAYMENT_FAILURE_TRANSIENT_ERROR_RETRY_LATER = exports.PAYMENT_FAILURE_PAYMENT_ALREADY_IN_FLIGHT = exports.PAYMENT_FAILURE_PAYMENT_ALREADY_SETTLED = exports.PAYMENT_FAILURE_INVOICE_EXPIRED = exports.PAYMENT_FAILURE_EXCESS_WALLET_BALANCE = exports.PAYMENT_FAILURE_INSUFFICIENT_WALLET_BALANCE = exports.PAYMENT_FAILURE_INSUFFICIENT_CHANNEL_BALANCE = exports.PAYMENT_FAILURE_INCORRECT_DETAILS = exports.PAYMENT_FAILURE_ERROR = exports.PAYMENT_FAILURE_NO_ROUTE = exports.PAYMENT_FAILURE_TIMEOUT = exports.PaymentStatus = void 0;
const t = __importStar(require("io-ts"));
const BigIntFromString_1 = require("io-ts-types/BigIntFromString");
const DateFromISOString_1 = require("io-ts-types/DateFromISOString");
const public_types_1 = require("@bitgo/public-types");
// codecs for lightning wallet payment related apis
exports.PaymentStatus = t.union([
    // Initial state when payment is initiated.
    // Transitions to 'settled' or 'failed' on LND notification.
    t.literal('in_flight'),
    // Final state.
    t.literal('settled'),
    // Final state.
    t.literal('failed'),
]);
exports.PAYMENT_FAILURE_TIMEOUT = 'TIMEOUT';
exports.PAYMENT_FAILURE_NO_ROUTE = 'NO_ROUTE';
/** A non-recoverable error */
exports.PAYMENT_FAILURE_ERROR = 'ERROR';
exports.PAYMENT_FAILURE_INCORRECT_DETAILS = 'INCORRECT_PAYMENT_DETAILS';
/** Insufficient channel outbound capacity */
exports.PAYMENT_FAILURE_INSUFFICIENT_CHANNEL_BALANCE = 'INSUFFICIENT_BALANCE';
/** Insufficient custodial lightning balance for the customer's wallet */
exports.PAYMENT_FAILURE_INSUFFICIENT_WALLET_BALANCE = 'INSUFFICIENT_WALLET_BALANCE';
/** Excess custodial lightning balance for the customer's wallet */
exports.PAYMENT_FAILURE_EXCESS_WALLET_BALANCE = 'EXCESS_WALLET_BALANCE';
exports.PAYMENT_FAILURE_INVOICE_EXPIRED = 'INVOICE_EXPIRED';
exports.PAYMENT_FAILURE_PAYMENT_ALREADY_SETTLED = 'PAYMENT_ALREADY_SETTLED';
exports.PAYMENT_FAILURE_PAYMENT_ALREADY_IN_FLIGHT = 'PAYMENT_ALREADY_IN_FLIGHT';
exports.PAYMENT_FAILURE_TRANSIENT_ERROR_RETRY_LATER = 'TRANSIENT_ERROR_RETRY_LATER';
exports.PAYMENT_FAILURE_CANCELED = 'CANCELED';
exports.PAYMENT_FAILURE_FORCE_FAILED = 'FORCE_FAILED';
exports.PaymentFailureReason = t.union([
    t.literal(exports.PAYMENT_FAILURE_TIMEOUT),
    t.literal(exports.PAYMENT_FAILURE_NO_ROUTE),
    t.literal(exports.PAYMENT_FAILURE_ERROR),
    t.literal(exports.PAYMENT_FAILURE_INCORRECT_DETAILS),
    t.literal(exports.PAYMENT_FAILURE_INSUFFICIENT_CHANNEL_BALANCE),
    t.literal(exports.PAYMENT_FAILURE_INSUFFICIENT_WALLET_BALANCE),
    t.literal(exports.PAYMENT_FAILURE_EXCESS_WALLET_BALANCE),
    t.literal(exports.PAYMENT_FAILURE_INVOICE_EXPIRED),
    t.literal(exports.PAYMENT_FAILURE_PAYMENT_ALREADY_SETTLED),
    t.literal(exports.PAYMENT_FAILURE_PAYMENT_ALREADY_IN_FLIGHT),
    t.literal(exports.PAYMENT_FAILURE_TRANSIENT_ERROR_RETRY_LATER),
    t.literal(exports.PAYMENT_FAILURE_CANCELED),
    t.literal(exports.PAYMENT_FAILURE_FORCE_FAILED),
]);
/**
 * Off-chain payment information
 */
exports.PaymentInfo = t.intersection([
    t.type({
        id: t.string,
        paymentHash: t.string,
        walletId: t.string,
        txRequestId: t.string,
        status: exports.PaymentStatus,
        invoice: t.string,
        feeLimitMsat: BigIntFromString_1.BigIntFromString,
        destination: t.string,
        updatedAt: DateFromISOString_1.DateFromISOString,
        createdAt: DateFromISOString_1.DateFromISOString,
        amountMsat: BigIntFromString_1.BigIntFromString,
    }),
    t.partial({
        feeMsat: BigIntFromString_1.BigIntFromString,
        failureReason: exports.PaymentFailureReason,
        paymentPreimage: t.string,
    }),
], 'PaymentInfo');
exports.ListPaymentsResponse = t.intersection([
    t.type({
        payments: t.array(exports.PaymentInfo),
    }),
    t.partial({
        /**
         * This is the paymentId of the last Payment in the last iteration.
         * Providing this value as the prevId in the next request will return the next batch of payments.
         * */
        nextBatchPrevId: t.string,
    }),
], 'ListPaymentsResponse');
/**
 * Payment query parameters
 */
exports.PaymentQuery = t.partial({
    status: exports.PaymentStatus,
    limit: BigIntFromString_1.BigIntFromString,
    startDate: DateFromISOString_1.DateFromISOString,
    endDate: DateFromISOString_1.DateFromISOString,
    paymentHash: t.string,
    /** paymentId provided by nextBatchPrevId in the previous list */
    prevId: t.string,
}, 'PaymentQuery');
exports.SubmitPaymentParams = t.intersection([
    public_types_1.LightningPaymentRequest,
    t.type({
        passphrase: t.string,
    }),
    t.partial({
        sequenceId: public_types_1.optionalString,
        comment: public_types_1.optionalString,
    }),
]);
exports.LndCreatePaymentResponse = t.intersection([
    t.type({
        status: exports.PaymentStatus,
        paymentHash: t.string,
    }),
    t.partial({
        paymentId: t.string,
        paymentPreimage: t.string,
        amountMsat: t.string,
        feeMsat: t.string,
        failureReason: exports.PaymentFailureReason,
    }),
], 'LndCreatePaymentResponse');
