// @bitgo/abstract-lightning/dist/src/codecs/api/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionQuery = exports.ListTransactionsResponse = exports.Transaction = exports.Output = exports.TransactionEntry = void 0;
const t = __importStar(require("io-ts"));
const DateFromISOString_1 = require("io-ts-types/DateFromISOString");
const BigIntFromString_1 = require("io-ts-types/BigIntFromString");
// codecs for lightning wallet transaction related apis
/**
 * Transaction entry details
 */
exports.TransactionEntry = t.intersection([
    t.type({
        inputs: t.number,
        outputs: t.number,
        value: t.number,
        valueString: t.string,
        address: t.string,
    }),
    t.partial({
        wallet: t.string,
    }),
], 'TransactionEntry');
/**
 * Transaction input/output details
 */
exports.Output = t.intersection([
    t.type({
        id: t.string,
        value: t.number,
        valueString: t.string,
    }),
    t.partial({
        wallet: t.string,
        address: t.string,
    }),
], 'Output');
/**
 * Transaction details
 */
exports.Transaction = t.intersection([
    t.type({
        id: t.string,
        normalizedTxHash: t.string,
        blockHeight: t.number,
        inputIds: t.array(t.string),
        entries: t.array(exports.TransactionEntry),
        inputs: t.array(exports.Output),
        outputs: t.array(exports.Output),
        size: t.number,
        date: DateFromISOString_1.DateFromISOString,
        fee: t.number,
        feeString: t.string,
        hex: t.string,
        confirmations: t.number,
    }),
    t.partial({
        blockHash: t.string,
        blockPosition: t.number,
        label: t.string,
    }),
], 'Transaction');
exports.ListTransactionsResponse = t.intersection([
    t.type({
        transactions: t.array(exports.Transaction),
    }),
    t.partial({
        /**
         * Transaction ID of the last transaction in this batch.
         * Use as prevId in next request to continue pagination.
         */
        nextBatchPrevId: t.string,
    }),
], 'ListTransactionsResponse');
/**
 * Transaction query parameters with cursor-based pagination
 */
exports.TransactionQuery = t.partial({
    /** Maximum number of transactions to return per page */
    limit: BigIntFromString_1.BigIntFromString,
    /** Optional filter for transactions at a specific block height */
    blockHeight: BigIntFromString_1.BigIntFromString,
    /** Optional start date filter */
    startDate: DateFromISOString_1.DateFromISOString,
    /** Optional end date filter */
    endDate: DateFromISOString_1.DateFromISOString,
    /** Transaction ID for cursor-based pagination (from nextBatchPrevId) */
    prevId: t.string,
}, 'TransactionQuery');
