// @bitgo/abstract-lightning/dist/src/codecs/api/wallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateLightningWalletClientRequest = exports.UpdateLightningWalletEncryptedRequest = exports.WatchOnly = exports.WatchOnlyAccount = exports.LightningAuthKeychain = exports.LightningKeychain = exports.LightningAuthKeychainCoinSpecific = exports.KeyPurpose = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const shared_1 = require("../shared");
// codec for lightning wallet and keychain related apis
exports.KeyPurpose = t.union([t.literal('userAuth'), t.literal('nodeAuth')], 'KeyPurpose');
exports.LightningAuthKeychainCoinSpecific = (0, shared_1.getCodecPair)(t.type({ purpose: exports.KeyPurpose }));
exports.LightningKeychain = t.intersection([
    t.type({
        id: io_ts_types_1.NonEmptyString,
        pub: io_ts_types_1.NonEmptyString,
        source: t.literal('user'),
    }),
    t.partial({
        encryptedPrv: io_ts_types_1.NonEmptyString,
    }),
], 'LightningKeychain');
exports.LightningAuthKeychain = t.intersection([
    t.type({
        id: io_ts_types_1.NonEmptyString,
        pub: io_ts_types_1.NonEmptyString,
        coinSpecific: exports.LightningAuthKeychainCoinSpecific,
        source: t.literal('user'),
    }),
    t.partial({
        encryptedPrv: io_ts_types_1.NonEmptyString,
    }),
], 'LightningAuthKeychain');
exports.WatchOnlyAccount = t.type({
    purpose: t.number,
    coin_type: t.number,
    account: t.number,
    xpub: t.string,
});
exports.WatchOnly = t.intersection([
    t.type({
        accounts: t.array(exports.WatchOnlyAccount),
    }),
    t.partial({
        master_key_birthday_timestamp: t.string,
        master_key_fingerprint: t.string,
    }),
]);
const CommonLightningUpdateWalletFields = t.partial({
    signerHost: t.string,
    signerTlsCert: t.string,
    watchOnlyAccounts: exports.WatchOnly,
});
exports.UpdateLightningWalletEncryptedRequest = t.intersection([
    CommonLightningUpdateWalletFields,
    t.partial({
        encryptedSignerMacaroon: t.string,
        encryptedSignerAdminMacaroon: t.string,
        encryptedSignerTlsKey: t.string,
    }),
]);
exports.UpdateLightningWalletClientRequest = t.intersection([
    CommonLightningUpdateWalletFields,
    t.type({
        passphrase: t.string,
    }),
    t.partial({
        signerMacaroon: t.string,
        signerAdminMacaroon: t.string,
        signerTlsKey: t.string,
    }),
]);
