// @bitgo/abstract-lightning/dist/src/codecs/api/withdraw.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendPsbtResponse = exports.SendPsbtRequest = exports.SignPsbtResponse = exports.SignPsbtRequest = exports.FundPsbtResponse = exports.LndCreateWithdrawResponse = exports.LightningOnchainWithdrawParams = exports.WithdrawStatus = exports.WithdrawStatusFailed = exports.WithdrawStatusDelivered = void 0;
const t = __importStar(require("io-ts"));
const public_types_1 = require("@bitgo/public-types");
exports.WithdrawStatusDelivered = 'delivered';
exports.WithdrawStatusFailed = 'failed';
exports.WithdrawStatus = t.union([t.literal(exports.WithdrawStatusDelivered), t.literal(exports.WithdrawStatusFailed)]);
exports.LightningOnchainWithdrawParams = t.intersection([
    public_types_1.LightningOnchainRequest,
    t.type({
        passphrase: t.string,
    }),
    t.partial({
        sequenceId: public_types_1.optionalString,
        comment: public_types_1.optionalString,
    }),
]);
exports.LndCreateWithdrawResponse = t.intersection([
    t.type({
        status: exports.WithdrawStatus,
    }),
    t.partial({
        txid: t.string,
        failureReason: t.string,
    }),
], 'LndCreateWithdrawResponse');
exports.FundPsbtResponse = t.type({
    fundedPsbt: t.string,
    changeOutputIndex: t.number,
}, 'FundPsbtResponse');
exports.SignPsbtRequest = t.type({
    fundedPsbt: t.string,
    txRequestId: t.string,
    userAuthKey: t.string,
    signature: t.string,
    signedRequest: t.string,
}, 'SendPsbtRequest');
exports.SignPsbtResponse = t.type({
    signedPsbt: t.string,
    signedTxHex: t.string,
}, 'SendPsbtResponse');
exports.SendPsbtRequest = t.type({
    signedTxHex: t.string,
    txRequestId: t.string,
    userAuthKey: t.string,
    signature: t.string,
    signedRequest: t.string,
}, 'SendPsbtRequest');
exports.SendPsbtResponse = t.intersection([
    t.type({
        label: t.string,
    }),
    t.partial({
        status: exports.WithdrawStatus,
        failureReason: t.string,
    }),
], 'SendPsbtResponse');
