// @bitgo/sdk-coin-flrp/dist/test/unit/lib/exportInPTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
require("should");
const exportInP_1 = require("../../resources/transactionData/exportInP");
const account_1 = require("../../resources/account");
const lib_1 = require("../../../src/lib");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const statics_1 = require("@bitgo/statics");
const signFlowTestSuit_1 = __importDefault(require("./signFlowTestSuit"));
describe('Flrp Export In P Tx Builder', () => {
    const coinConfig = statics_1.coins.get('tflrp');
    const factory = new lib_1.TransactionBuilderFactory(coinConfig);
    describe('validate txBuilder fields', () => {
        const txBuilder = factory.getExportInPBuilder();
        it('should fail amount low than zero', () => {
            assert_1.default.throws(() => {
                txBuilder.amount('-1');
            }, (e) => e.message === 'Amount must be greater than 0');
        });
        it('should fail target chain id length incorrect', () => {
            assert_1.default.throws(() => {
                txBuilder.externalChainId(Buffer.from(exportInP_1.EXPORT_IN_P.INVALID_CHAIN_ID.slice(2)));
            }, (e) => e.message === 'Chain id are 32 byte size');
        });
        it('should fail target chain id not a valid base58 string', () => {
            assert_1.default.throws(() => {
                txBuilder.externalChainId(exportInP_1.EXPORT_IN_P.INVALID_CHAIN_ID);
            }, (e) => e.message === 'Non-base58 character');
        });
        it('should fail target chain id cb58 invalid checksum', () => {
            assert_1.default.throws(() => {
                txBuilder.externalChainId(exportInP_1.EXPORT_IN_P.VALID_C_CHAIN_ID.slice(2));
            }, (e) => e.message === 'Invalid checksum');
        });
        it('should fail validate Utxos empty string', () => {
            assert_1.default.throws(() => {
                txBuilder.validateUtxos([]);
            }, (e) => e.message === 'UTXOs array cannot be empty');
        });
        it('should throw if feeState is not set', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .amount('500000000')
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId);
            await txBuilder.build().should.be.rejectedWith('Fee state is required');
        });
        it('should accept valid feeState', () => {
            const txBuilder = factory.getExportInPBuilder();
            (() => txBuilder.feeState(exportInP_1.EXPORT_IN_P.feeState)).should.not.throw();
        });
        it('should throw if context is not set', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .amount('500000000')
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .feeState(exportInP_1.EXPORT_IN_P.feeState)
                .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos);
            // context is NOT set
            await txBuilder.build().should.be.rejectedWith('context is required');
        });
        it('should fail when utxos hex array is empty', () => {
            const txBuilder = factory.getExportInPBuilder();
            assert_1.default.throws(() => {
                txBuilder.decodedUtxos([]);
            }, (e) => e.message === 'UTXOs array cannot be empty');
        });
        it('should throw if amount is not set', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .feeState(exportInP_1.EXPORT_IN_P.feeState)
                .context(exportInP_1.EXPORT_IN_P.context)
                .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos);
            // amount is NOT set
            await txBuilder.build().should.be.rejectedWith('amount is required');
        });
    });
    (0, signFlowTestSuit_1.default)({
        transactionType: 'Export P2C with changeoutput',
        newTxFactory: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tflrp')),
        newTxBuilder: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tflrp'))
            .getExportInPBuilder()
            .threshold(exportInP_1.EXPORT_IN_P.threshold)
            .locktime(exportInP_1.EXPORT_IN_P.locktime)
            .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
            .amount(exportInP_1.EXPORT_IN_P.amount)
            .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
            .feeState(exportInP_1.EXPORT_IN_P.feeState)
            .context(exportInP_1.EXPORT_IN_P.context)
            .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos),
        unsignedTxHex: exportInP_1.EXPORT_IN_P.unsignedHex,
        halfSignedTxHex: exportInP_1.EXPORT_IN_P.halfSigntxHex,
        fullSignedTxHex: exportInP_1.EXPORT_IN_P.fullSigntxHex,
        privateKey: {
            prv1: exportInP_1.EXPORT_IN_P.privateKeys[2],
            prv2: exportInP_1.EXPORT_IN_P.privateKeys[0],
        },
        txHash: exportInP_1.EXPORT_IN_P.txhash,
    });
    it('Should full sign a export tx from unsigned raw tx', () => {
        const txBuilder = new lib_1.TransactionBuilderFactory(statics_1.coins.get('tflrp')).from(exportInP_1.EXPORT_IN_P.unsignedHex);
        txBuilder.sign({ key: exportInP_1.EXPORT_IN_P.privateKeys[0] });
        txBuilder
            .build()
            .then(() => assert_1.default.fail('it can sign'))
            .catch((err) => {
            err.message.should.be.equal('Private key cannot sign the transaction');
        });
    });
    describe('Change output threshold fix', () => {
        /**
         * This test suite verifies the fix for the change output threshold bug.
         *
         * The issue: FlareJS's pvm.e.newExportTx() defaults change outputs to threshold=1,
         * but for multisig wallets we need threshold=2 to maintain proper security.
         *
         * The fix: After building the transaction, we correct the change outputs to use
         * the wallet's threshold (typically 2 for multisig).
         */
        it('should create change output with threshold=2 for multisig wallets', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .amount(exportInP_1.EXPORT_IN_P.amount)
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .feeState(exportInP_1.EXPORT_IN_P.feeState)
                .context(exportInP_1.EXPORT_IN_P.context)
                .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos);
            const tx = await txBuilder.build();
            const flareTransaction = tx._flareTransaction;
            const innerTx = flareTransaction.getTx();
            const changeOutputs = innerTx.baseTx.outputs;
            changeOutputs.length.should.be.greaterThan(0);
            changeOutputs.forEach((output, index) => {
                const transferOut = output.output;
                const threshold = transferOut.outputOwners.threshold.value();
                threshold.should.equal(exportInP_1.EXPORT_IN_P.threshold, `Change output ${index} should have threshold=${exportInP_1.EXPORT_IN_P.threshold}, but has threshold=${threshold}`);
            });
        });
        it('should create change output with correct locktime for multisig wallets', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .amount(exportInP_1.EXPORT_IN_P.amount)
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .feeState(exportInP_1.EXPORT_IN_P.feeState)
                .context(exportInP_1.EXPORT_IN_P.context)
                .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos);
            const tx = await txBuilder.build();
            const flareTransaction = tx._flareTransaction;
            const innerTx = flareTransaction.getTx();
            const changeOutputs = innerTx.baseTx.outputs;
            changeOutputs.length.should.be.greaterThan(0);
            changeOutputs.forEach((output, index) => {
                const transferOut = output.output;
                const locktime = transferOut.outputOwners.locktime.value();
                locktime.should.equal(BigInt(exportInP_1.EXPORT_IN_P.locktime), `Change output ${index} should have locktime=${exportInP_1.EXPORT_IN_P.locktime}, but has locktime=${locktime}`);
            });
        });
        it('should maintain threshold=2 after parsing and rebuilding', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .amount(exportInP_1.EXPORT_IN_P.amount)
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .feeState(exportInP_1.EXPORT_IN_P.feeState)
                .context(exportInP_1.EXPORT_IN_P.context)
                .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos);
            const tx = await txBuilder.build();
            const txHex = tx.toBroadcastFormat();
            const parsedTxBuilder = factory.from(txHex);
            const parsedTx = await parsedTxBuilder.build();
            const flareTransaction = parsedTx._flareTransaction;
            const innerTx = flareTransaction.getTx();
            const changeOutputs = innerTx.baseTx.outputs;
            changeOutputs.length.should.be.greaterThan(0);
            changeOutputs.forEach((output, index) => {
                const transferOut = output.output;
                const threshold = transferOut.outputOwners.threshold.value();
                threshold.should.equal(exportInP_1.EXPORT_IN_P.threshold, `After parsing, change output ${index} should have threshold=${exportInP_1.EXPORT_IN_P.threshold}, but has threshold=${threshold}`);
            });
        });
        it('should have change output addresses matching wallet addresses', async () => {
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(exportInP_1.EXPORT_IN_P.threshold)
                .locktime(exportInP_1.EXPORT_IN_P.locktime)
                .fromPubKey(exportInP_1.EXPORT_IN_P.pAddresses)
                .amount(exportInP_1.EXPORT_IN_P.amount)
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .feeState(exportInP_1.EXPORT_IN_P.feeState)
                .context(exportInP_1.EXPORT_IN_P.context)
                .decodedUtxos(exportInP_1.EXPORT_IN_P.utxos);
            const tx = await txBuilder.build();
            const flareTransaction = tx._flareTransaction;
            const innerTx = flareTransaction.getTx();
            const changeOutputs = innerTx.baseTx.outputs;
            changeOutputs.length.should.be.greaterThan(0);
            changeOutputs.forEach((output) => {
                const transferOut = output.output;
                const addresses = transferOut.outputOwners.addrs;
                addresses.length.should.equal(3, 'Change output should have 3 addresses for multisig wallet');
            });
        });
    });
    describe('on-chain verified transactions', () => {
        it('should build and sign export tx with correct sigIndices - on-chain verified', async () => {
            const utxos = [
                {
                    outputID: 7,
                    amount: '50000000',
                    txid: 'bgHnEJ64td8u31aZrGDaWcDqxZ8vDV5qGd7bmSifgvUnUW8v2',
                    threshold: 2,
                    addresses: [
                        account_1.ON_CHAIN_TEST_WALLET.bitgo.pChainAddress,
                        account_1.ON_CHAIN_TEST_WALLET.backup.pChainAddress,
                        account_1.ON_CHAIN_TEST_WALLET.user.pChainAddress,
                    ],
                    outputidx: '0',
                    locktime: '0',
                },
                {
                    outputID: 7,
                    amount: '50000000',
                    txid: 'KdrKz1SHM11dpDGHUthRc9sgS1hnb48pfvnmZDtJu7dRFF2Ha',
                    threshold: 2,
                    addresses: [
                        account_1.ON_CHAIN_TEST_WALLET.bitgo.pChainAddress,
                        account_1.ON_CHAIN_TEST_WALLET.backup.pChainAddress,
                        account_1.ON_CHAIN_TEST_WALLET.user.pChainAddress,
                    ],
                    outputidx: '0',
                    locktime: '0',
                },
            ];
            const senderPAddresses = [
                account_1.ON_CHAIN_TEST_WALLET.user.pChainAddress,
                account_1.ON_CHAIN_TEST_WALLET.bitgo.pChainAddress,
                account_1.ON_CHAIN_TEST_WALLET.backup.pChainAddress,
            ];
            const exportAmount = '30000000';
            const txBuilder = factory
                .getExportInPBuilder()
                .threshold(2)
                .locktime(0)
                .fromPubKey(senderPAddresses)
                .amount(exportAmount)
                .externalChainId(exportInP_1.EXPORT_IN_P.sourceChainId)
                .decodedUtxos(utxos)
                .context(exportInP_1.EXPORT_IN_P.context)
                .feeState(exportInP_1.EXPORT_IN_P.feeState);
            txBuilder.sign({ key: account_1.ON_CHAIN_TEST_WALLET.user.privateKey });
            txBuilder.sign({ key: account_1.ON_CHAIN_TEST_WALLET.bitgo.privateKey });
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            tx.id.should.equal('nSBwNcgfLbk5S425b1qaYaqTTCiMCV75KU4Fbnq8SPUUqLq2');
            const hex = rawTx.replace('0x', '');
            const amountHex = '0000000002faf080';
            const amountPos = hex.indexOf(amountHex);
            amountPos.should.be.greaterThan(0);
            const inputSection = hex.substring(amountPos + 16, amountPos + 40);
            const numSigIndices = parseInt(inputSection.substring(0, 8), 16);
            const sigIdx0 = parseInt(inputSection.substring(8, 16), 16);
            const sigIdx1 = parseInt(inputSection.substring(16, 24), 16);
            numSigIndices.should.equal(2);
            sigIdx0.should.equal(0);
            sigIdx1.should.equal(2);
            const flareTransaction = tx._flareTransaction;
            const innerTx = flareTransaction.getTx();
            const changeOutputs = innerTx.baseTx.outputs;
            changeOutputs.length.should.be.greaterThan(0, 'Should have change output');
            const changeOutput = changeOutputs[0].output;
            changeOutput.outputOwners.threshold.value().should.equal(2);
            changeOutput.outputOwners.addrs.length.should.equal(3);
            const expectedAddressBytes = senderPAddresses.map((addr) => utils_1.default.parseAddress(addr));
            const expectedAddressHexes = expectedAddressBytes.map((buf) => buf.toString('hex')).sort();
            const actualAddressHexes = changeOutput.outputOwners.addrs.map((addr) => addr.toHex().replace('0x', '')).sort();
            actualAddressHexes.should.deepEqual(expectedAddressHexes);
        });
    });
});
