// @bitgo/sdk-coin-flrp/dist/test/unit/lib/importInCTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
require("should");
const lib_1 = require("../../../src/lib");
const statics_1 = require("@bitgo/statics");
const importInC_1 = require("../../resources/transactionData/importInC");
const account_1 = require("../../resources/account");
const signFlowTestSuit_1 = __importDefault(require("./signFlowTestSuit"));
describe('Flrp Import In C Tx Builder', () => {
    const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get('tflrp'));
    describe('validate txBuilder fields', () => {
        it('should fail validate Utxos empty array', () => {
            const txBuilder = factory.getImportInCBuilder();
            assert_1.default.throws(() => {
                txBuilder.validateUtxos([]);
            }, (e) => e.message === 'UTXOs array cannot be empty');
        });
        it('should throw if to address is not set', async () => {
            const txBuilder = factory
                .getImportInCBuilder()
                .threshold(importInC_1.IMPORT_IN_C.threshold)
                .fromPubKey(importInC_1.IMPORT_IN_C.pAddresses)
                .decodedUtxos(importInC_1.IMPORT_IN_C.utxos)
                .fee(importInC_1.IMPORT_IN_C.fee)
                .context(importInC_1.IMPORT_IN_C.context);
            // to is NOT set
            await txBuilder.build().should.be.rejectedWith('to is required');
        });
        it('should throw if context is not set', async () => {
            const txBuilder = factory
                .getImportInCBuilder()
                .threshold(importInC_1.IMPORT_IN_C.threshold)
                .fromPubKey(importInC_1.IMPORT_IN_C.pAddresses)
                .decodedUtxos(importInC_1.IMPORT_IN_C.utxos)
                .to(importInC_1.IMPORT_IN_C.to)
                .fee(importInC_1.IMPORT_IN_C.fee);
            // context is NOT set
            await txBuilder.build().should.be.rejectedWith('context is required');
        });
        it('should throw if fromAddresses is not set', async () => {
            const txBuilder = factory
                .getImportInCBuilder()
                .threshold(importInC_1.IMPORT_IN_C.threshold)
                .decodedUtxos(importInC_1.IMPORT_IN_C.utxos)
                .to(importInC_1.IMPORT_IN_C.to)
                .fee(importInC_1.IMPORT_IN_C.fee)
                .context(importInC_1.IMPORT_IN_C.context);
            // fromPubKey is NOT set
            await txBuilder.build().should.be.rejectedWith('fromAddresses are required');
        });
        it('should throw if UTXOs are not set', async () => {
            const txBuilder = factory
                .getImportInCBuilder()
                .threshold(importInC_1.IMPORT_IN_C.threshold)
                .fromPubKey(importInC_1.IMPORT_IN_C.pAddresses)
                .to(importInC_1.IMPORT_IN_C.to)
                .fee(importInC_1.IMPORT_IN_C.fee)
                .context(importInC_1.IMPORT_IN_C.context);
            // utxos is NOT set
            await txBuilder.build().should.be.rejectedWith('UTXOs are required');
        });
        it('should fail when utxos hex array is empty', () => {
            const txBuilder = factory.getImportInCBuilder();
            assert_1.default.throws(() => {
                txBuilder.decodedUtxos([]);
            }, (e) => e.message === 'UTXOs array cannot be empty');
        });
        it('should fail with invalid threshold value (0)', () => {
            const txBuilder = factory.getImportInCBuilder();
            assert_1.default.throws(() => {
                txBuilder.threshold(0);
            }, (e) => e.message.includes('threshold') || e.message.includes('greater'));
        });
        it('should fail with invalid threshold value (negative)', () => {
            const txBuilder = factory.getImportInCBuilder();
            assert_1.default.throws(() => {
                txBuilder.threshold(-1);
            }, (e) => e.message.includes('threshold') || e.message.includes('greater'));
        });
        it('should fail with invalid to address format', () => {
            const txBuilder = factory.getImportInCBuilder();
            assert_1.default.throws(() => {
                txBuilder.to('invalid-address');
            }, (e) => e.message.includes('Invalid') || e.message.includes('address'));
        });
        it('should accept valid to address', () => {
            const txBuilder = factory.getImportInCBuilder();
            (() => txBuilder.to(importInC_1.IMPORT_IN_C.to)).should.not.throw();
        });
        it('should accept valid threshold', () => {
            const txBuilder = factory.getImportInCBuilder();
            (() => txBuilder.threshold(2)).should.not.throw();
        });
        it('should accept valid context', () => {
            const txBuilder = factory.getImportInCBuilder();
            (() => txBuilder.context(importInC_1.IMPORT_IN_C.context)).should.not.throw();
        });
        it('should accept valid fromPubKey addresses', () => {
            const txBuilder = factory.getImportInCBuilder();
            (() => txBuilder.fromPubKey(importInC_1.IMPORT_IN_C.pAddresses)).should.not.throw();
        });
    });
    (0, signFlowTestSuit_1.default)({
        transactionType: 'Import C2P',
        newTxFactory: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tflrp')),
        newTxBuilder: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tflrp'))
            .getImportInCBuilder()
            .threshold(importInC_1.IMPORT_IN_C.threshold)
            .fromPubKey(importInC_1.IMPORT_IN_C.pAddresses)
            .decodedUtxos(importInC_1.IMPORT_IN_C.utxos)
            .to(importInC_1.IMPORT_IN_C.to)
            .fee(importInC_1.IMPORT_IN_C.fee)
            .context(importInC_1.IMPORT_IN_C.context),
        unsignedTxHex: importInC_1.IMPORT_IN_C.unsignedHex,
        halfSignedTxHex: importInC_1.IMPORT_IN_C.halfSigntxHex,
        fullSignedTxHex: importInC_1.IMPORT_IN_C.fullSigntxHex,
        privateKey: {
            prv1: importInC_1.IMPORT_IN_C.privateKeys[2],
            prv2: importInC_1.IMPORT_IN_C.privateKeys[0],
        },
        txHash: importInC_1.IMPORT_IN_C.txhash,
    });
    describe('on-chain verified transactions', () => {
        it('should build and sign import to C-chain tx with correct sigIndices - on-chain verified', async () => {
            const utxo = {
                outputID: 7,
                amount: '30000000',
                txid: 'nSBwNcgfLbk5S425b1qaYaqTTCiMCV75KU4Fbnq8SPUUqLq2',
                threshold: 2,
                addresses: [
                    account_1.ON_CHAIN_TEST_WALLET.bitgo.pChainAddress,
                    account_1.ON_CHAIN_TEST_WALLET.backup.pChainAddress,
                    account_1.ON_CHAIN_TEST_WALLET.user.pChainAddress,
                ],
                outputidx: '1',
                locktime: '0',
            };
            const senderPAddresses = [
                account_1.ON_CHAIN_TEST_WALLET.user.pChainAddress,
                account_1.ON_CHAIN_TEST_WALLET.bitgo.pChainAddress,
                account_1.ON_CHAIN_TEST_WALLET.backup.pChainAddress,
            ];
            const toAddress = '0x96993BAEb6AaE2e06BF95F144e2775D4f8efbD35';
            const importFee = '1000000';
            const txBuilder = factory
                .getImportInCBuilder()
                .threshold(2)
                .locktime(0)
                .fromPubKey(senderPAddresses)
                .to(toAddress)
                .fee(importFee)
                .externalChainId(importInC_1.IMPORT_IN_C.sourceChainId)
                .decodedUtxos([utxo])
                .context(importInC_1.IMPORT_IN_C.context);
            txBuilder.sign({ key: account_1.ON_CHAIN_TEST_WALLET.user.privateKey });
            txBuilder.sign({ key: account_1.ON_CHAIN_TEST_WALLET.bitgo.privateKey });
            const tx = await txBuilder.build();
            const rawTx = tx.toBroadcastFormat();
            tx.id.should.equal('2t4gxEAdPLiiy9HsbjaQun1mVFewMoixNS64eZ56C38L4mpP1j');
            const hex = rawTx.replace('0x', '');
            const amountHex = '0000000001c9c380';
            const amountPos = hex.indexOf(amountHex);
            amountPos.should.be.greaterThan(0);
            const inputSection = hex.substring(amountPos + 16, amountPos + 40);
            const numSigIndices = parseInt(inputSection.substring(0, 8), 16);
            const sigIdx0 = parseInt(inputSection.substring(8, 16), 16);
            const sigIdx1 = parseInt(inputSection.substring(16, 24), 16);
            numSigIndices.should.equal(2);
            sigIdx0.should.equal(0);
            sigIdx1.should.equal(2);
            const expectedOutput = parseInt(utxo.amount, 10) - parseInt(importFee, 10);
            const outputHex = expectedOutput.toString(16).padStart(16, '0');
            hex.should.containEql(outputHex);
        });
    });
});
