// @bitgo/sdk-coin-flrp/dist/test/unit/lib/permissionlessDelegatorTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
require("should");
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const lib_1 = require("../../../src/lib");
const account_1 = require("../../resources/account");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const delegatorTx_1 = require("../../resources/transactionData/delegatorTx");
describe('Flrp PermissionlessDelegatorTxBuilder', () => {
    const coinConfig = statics_1.coins.get('tflrp');
    const factory = new lib_1.TransactionBuilderFactory(coinConfig);
    describe('getDelegatorBuilder', () => {
        it('should return a PermissionlessDelegatorTxBuilder instance', () => {
            const builder = factory.getDelegatorBuilder();
            builder.should.be.instanceOf(lib_1.PermissionlessDelegatorTxBuilder);
        });
    });
    describe('validate nodeID', () => {
        it('should fail when nodeID is missing', () => {
            const builder = factory.getDelegatorBuilder();
            assert_1.default.throws(() => {
                builder.nodeID('');
            }, (e) => e.message === 'Invalid transaction: missing nodeID');
        });
        it('should fail when nodeID has invalid prefix', () => {
            const builder = factory.getDelegatorBuilder();
            assert_1.default.throws(() => {
                builder.nodeID('InvalidPrefix-123456789');
            }, (e) => e.message === 'Invalid transaction: invalid NodeID tag');
        });
        it('should accept valid nodeID format', () => {
            const builder = factory.getDelegatorBuilder();
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            (() => builder.nodeID(validNodeID)).should.not.throw();
        });
    });
    describe('validate stake duration', () => {
        it('should fail when end time is less than start time', () => {
            const builder = factory.getDelegatorBuilder();
            const startTime = BigInt(Math.floor(Date.now() / 1000) + 60);
            const endTime = BigInt(Math.floor(Date.now() / 1000) - 60);
            builder.startTime(startTime.toString());
            builder.endTime(endTime.toString());
            // The validation happens in buildFlareTransaction, but we can test validateStakeDuration directly
            assert_1.default.throws(() => {
                builder.validateStakeDuration(startTime, endTime);
            }, (e) => e.message === 'End date cannot be less than start date');
        });
        it('should accept valid stake duration', () => {
            const builder = factory.getDelegatorBuilder();
            const startTime = BigInt(Math.floor(Date.now() / 1000) + 60);
            const endTime = BigInt(Math.floor(Date.now() / 1000) + 86400 * 14); // 14 days later
            (() => {
                builder.validateStakeDuration(startTime, endTime);
            }).should.not.throw();
        });
    });
    describe('builder methods', () => {
        it('should set startTime correctly', () => {
            const builder = factory.getDelegatorBuilder();
            const startTime = Math.floor(Date.now() / 1000) + 60;
            (() => builder.startTime(startTime)).should.not.throw();
        });
        it('should set endTime correctly', () => {
            const builder = factory.getDelegatorBuilder();
            const endTime = Math.floor(Date.now() / 1000) + 86400 * 14;
            (() => builder.endTime(endTime)).should.not.throw();
        });
        it('should set rewardAddress with P-chain address', () => {
            const builder = factory.getDelegatorBuilder();
            // P-chain Bech32 address
            (() => builder.rewardAddress(account_1.SEED_ACCOUNT.addressTestnet)).should.not.throw();
        });
    });
    describe('build validation', () => {
        it('should fail build when nodeID is not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            builder
                .startTime(now + 60)
                .endTime(now + 86400 * 14)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet])
                .context(account_1.CONTEXT);
            await builder.build().should.be.rejectedWith('NodeID is required for delegation');
        });
        it('should fail build when start time is not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .endTime(now + 86400 * 14)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet])
                .context(account_1.CONTEXT);
            await builder.build().should.be.rejectedWith('Start time is required for delegation');
        });
        it('should fail build when end time is not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet])
                .context(account_1.CONTEXT);
            await builder.build().should.be.rejectedWith('End time is required for delegation');
        });
        it('should fail build when stake amount is not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 86400 * 14)
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet])
                .context(account_1.CONTEXT);
            await builder.build().should.be.rejectedWith('Stake amount is required for delegation');
        });
        it('should fail build when context is not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 86400 * 14)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet]);
            await builder.build().should.be.rejectedWith('Context is required for delegation');
        });
        it('should fail build when from addresses are not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 86400 * 14)
                .stakeAmount(BigInt(50000 * 1e9))
                .context(account_1.CONTEXT);
            await builder.build().should.be.rejectedWith('From addresses are required for delegation');
        });
        it('should fail build when fee state is not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 86400 * 14)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet])
                .context(account_1.CONTEXT);
            await builder.build().should.be.rejectedWith('Fee state is required for delegation');
        });
        it('should fail build when UTXOs are not set', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const mockFeeState = { capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' };
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 86400 * 14)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey([account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet])
                .context(account_1.CONTEXT)
                .feeState(mockFeeState);
            await builder.build().should.be.rejectedWith('UTXOs are required for delegation');
        });
    });
    describe('validate stake amount', () => {
        it('should fail when stake amount is below minimum', () => {
            const builder = factory.getDelegatorBuilder();
            // Minimum is 50000 FLR, so 1000 should fail
            assert_1.default.throws(() => {
                builder.stakeAmount(BigInt(1000 * 1e9));
            }, (e) => e.message.includes('Minimum delegation amount'));
        });
        it('should accept stake amount at or above minimum', () => {
            const builder = factory.getDelegatorBuilder();
            const stakeAmount = BigInt(50000 * 1e9);
            (() => builder.stakeAmount(stakeAmount)).should.not.throw();
        });
    });
    describe('transaction type support', () => {
        it('should allow setting AddPermissionlessDelegator transaction type', () => {
            const builder = factory.getDelegatorBuilder();
            // The builder should successfully configure with AddPermissionlessDelegator type
            // This verifies transaction.ts setTransactionType accepts the delegation type
            builder.should.be.instanceof(lib_1.PermissionlessDelegatorTxBuilder);
        });
        it('should have correct minimum stake amount constant', () => {
            // Verify the minimum stake amount is 50000 FLR (in nanoFLR)
            const minStakeAmount = BigInt(50000 * 1e9);
            minStakeAmount.should.equal(BigInt('50000000000000'));
        });
    });
    describe('round-trip deserialization', () => {
        it('should deserialize a signed delegation transaction from hex', async () => {
            const rawTx = utils_1.default.removeHexPrefix(delegatorTx_1.DELEGATION_TX_2.signedHex);
            const tx = (await factory.from(rawTx).build());
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
        });
        it('should correctly extract delegation parameters from deserialized transaction', async () => {
            const rawTx = utils_1.default.removeHexPrefix(delegatorTx_1.DELEGATION_TX_2.signedHex);
            const tx = (await factory.from(rawTx).build());
            const explanation = tx.explainTransaction();
            explanation.outputs.length.should.be.above(0);
            explanation.outputAmount.should.equal(delegatorTx_1.DELEGATION_TX_2.stakeAmount);
        });
        it('should re-serialize to the same hex (round-trip)', async () => {
            const rawTx = utils_1.default.removeHexPrefix(delegatorTx_1.DELEGATION_TX_2.signedHex);
            const tx = (await factory.from(rawTx).build());
            const reEncodedTx = utils_1.default.removeHexPrefix(tx.toBroadcastFormat());
            reEncodedTx.should.equal(rawTx);
        });
        it('should have correct input/output counts', async () => {
            const rawTx = utils_1.default.removeHexPrefix(delegatorTx_1.DELEGATION_TX_2.signedHex);
            const tx = (await factory.from(rawTx).build());
            tx.inputs.should.have.length(3);
            tx.outputs.length.should.be.above(0);
        });
    });
    describe('build with UTXOs (comprehensive tests)', () => {
        const mockFeeState = { capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' };
        const MIN_DELEGATION_DURATION = 14 * 24 * 60 * 60; // 14 days (1209600 seconds)
        const MIN_DELEGATION_AMOUNT = BigInt(50000 * 1e9); // 50000 FLR in nanoFLR
        function createTestUtxos(addresses, threshold, amount = '60000000000000') {
            return [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount,
                    threshold,
                    addresses,
                    locktime: '0',
                },
            ];
        }
        it('should build delegation transaction with threshold=2 and 3 addresses (BitGo 2-of-3 multisig)', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = createTestUtxos(addresses, 2);
            const tx = await builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
        });
        it('should build delegation transaction with threshold=2 and 2 addresses', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet];
            const utxos = createTestUtxos(addresses, 2);
            const tx = await builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
        });
        it('should fail when UTXO amount is insufficient', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet];
            const utxos = createTestUtxos(addresses, 2, '1000000000000');
            await builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build()
                .should.be.rejected();
        });
        it('should build delegation with multiple UTXOs', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            // Create multiple UTXOs with different txids
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '30000000000000', // 30k FLR
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
                {
                    outputID: 7,
                    txid: '2kSSHWKZH7uJ1FfSpgRaPfgPpnT4915QVXPAMw6HjfFQVJ1QFx',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '25000000000000', // 25k FLR
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
                {
                    outputID: 7,
                    txid: '2E6Xuqf3i6TnwH6zjt4K6jNDR2ooj1DTJpTFZ6SgZZkRJ4ADSe',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '10000000000000', // 10k FLR
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            // Total: 65k FLR > 50k minimum
            const tx = await builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // FlareJS optimizes to use minimum UTXOs needed, may not consume all 3
            tx.inputs.length.should.be.greaterThan(0);
        });
        /**
         * Verifies fix for FlareJS Bug - Change output threshold
         *
         * FlareJS's pvm.e.newAddPermissionlessDelegatorTx() defaults change outputs to threshold=1.
         * We implement a fix (similar to ExportInPTxBuilder) to correct change outputs to threshold=2
         * for multisig wallets, maintaining security.
         */
        it('should create change output with threshold=2 for multisig wallet (bug fixed)', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = createTestUtxos(addresses, 2, '60000000000000'); // 60k FLR
            const tx = await builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT) // Stake 50k
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            // Verify change exists using explainTransaction
            const explanation = tx.explainTransaction();
            const changeAmount = BigInt(explanation.changeAmount);
            Number(changeAmount).should.be.greaterThan(Number(BigInt('9000000000000')), 'Should have change (at least 9k after fees)');
            explanation.changeOutputs.length.should.be.greaterThan(0, 'Should have change outputs');
            // Verify the fix: change outputs should now have threshold=2
            const flareTransaction = tx._flareTransaction;
            const innerTx = flareTransaction.getTx();
            const changeOutputs = innerTx.baseTx.outputs;
            changeOutputs.forEach((output) => {
                const transferOut = output.output;
                const threshold = transferOut.outputOwners.threshold.value();
                threshold.should.equal(2, 'Change output should have threshold=2 (bug fixed!)');
            });
        });
        it('should have change output addresses matching wallet addresses', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = createTestUtxos(addresses, 2, '60000000000000');
            const tx = await builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            // Verify change exists and addresses match using explainTransaction
            const explanation = tx.explainTransaction();
            const changeAmount = BigInt(explanation.changeAmount);
            Number(changeAmount).should.be.greaterThan(0, 'Should have change output');
            explanation.changeOutputs.length.should.be.greaterThan(0);
            // Verify change output addresses match wallet addresses (sorted)
            const sortedFromAddresses = tx.fromAddresses.slice().sort().join('~');
            explanation.changeOutputs.forEach((output) => {
                output.address.should.equal(sortedFromAddresses, 'Change output addresses should match wallet addresses');
            });
        });
    });
    describe('multiple delegations to same validator', () => {
        const mockFeeState = { capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' };
        const MIN_DELEGATION_DURATION = 14 * 24 * 60 * 60;
        const MIN_DELEGATION_AMOUNT = BigInt(50000 * 1e9);
        it('should support creating multiple delegations to same validator with different time periods', async () => {
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            // First delegation: Jan 1 - Mar 1
            const builder1 = factory.getDelegatorBuilder();
            const utxos1 = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '100000000000000', // 100k FLR
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            const tx1 = await builder1
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 60 + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT) // 50k
                .fromPubKey(addresses)
                .decodedUtxos(utxos1)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            tx1.should.be.instanceof(lib_1.Transaction);
            tx1.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Second delegation: Feb 1 - Apr 1 (overlapping period, different start/end)
            // Simulating using a different UTXO (in practice, could be change from first delegation)
            const builder2 = factory.getDelegatorBuilder();
            const utxos2 = [
                {
                    outputID: 7,
                    txid: '2QttuE1MNRPEvLPdhB8t6WpC91VKfvigHvf1PKrNQ6BGgB6hZw', // Valid CB58 txid
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '50010000000000', // 50010 FLR (50k stake + extra for fees)
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            const tx2 = await builder2
                .nodeID(validNodeID) // Same validator!
                .startTime(now + 60 + 30 * 24 * 60 * 60) // 30 days later
                .endTime(now + 60 + 30 * 24 * 60 * 60 + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT) // Another 50k
                .fromPubKey(addresses)
                .decodedUtxos(utxos2)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            tx2.should.be.instanceof(lib_1.Transaction);
            tx2.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Both transactions should be valid and independent
            tx1.id.should.not.equal(tx2.id);
        });
        it('should support sequential delegations (one ends, next begins)', async () => {
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '100000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            // First delegation
            const builder1 = factory.getDelegatorBuilder();
            const delegation1EndTime = now + 60 + MIN_DELEGATION_DURATION;
            const tx1 = await builder1
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(delegation1EndTime)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            // Second delegation starts right after first one ends
            const builder2 = factory.getDelegatorBuilder();
            const utxos2 = [
                {
                    outputID: 7,
                    txid: '2K9rgiEgsEcT3dB9EyduEHFq7g4sHQBMks16VUQArVRieBQ9W5', // Valid CB58 txid
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '50010000000000', // 50010 FLR (50k stake + extra for fees)
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            const tx2 = await builder2
                .nodeID(validNodeID)
                .startTime(delegation1EndTime + 1) // Starts 1 second after first ends
                .endTime(delegation1EndTime + 1 + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos2)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .build();
            tx1.should.be.instanceof(lib_1.Transaction);
            tx2.should.be.instanceof(lib_1.Transaction);
            tx1.id.should.not.equal(tx2.id);
        });
    });
    describe('signing multisig delegation transactions', () => {
        const mockFeeState = { capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' };
        const MIN_DELEGATION_DURATION = 14 * 24 * 60 * 60;
        const MIN_DELEGATION_AMOUNT = BigInt(50000 * 1e9);
        it('should successfully sign delegation transaction with threshold=2 multisig wallet', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000', // 60k FLR (will have change)
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT)
                .sign({ key: account_1.SEED_ACCOUNT.privateKey });
            const tx = await builder.build();
            // Verify transaction is built successfully
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify credentials exist (non-empty)
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            credentials.length.should.be.greaterThan(0);
            // Verify at least one signature slot has a signature
            let hasSignature = false;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    if (sig && sig.length > 0 && !sig.startsWith('0x' + '0'.repeat(130))) {
                        hasSignature = true;
                        break;
                    }
                }
                if (hasSignature)
                    break;
            }
            hasSignature.should.be.true('Should have at least one signature');
        });
        it('should successfully sign delegation transaction with multiple signers', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            // Sign with both keys for 2-of-2 multisig
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT);
            builder.sign({ key: account_1.SEED_ACCOUNT.privateKey });
            builder.sign({ key: account_1.ACCOUNT_1.privateKey });
            const tx = await builder.build();
            // Verify transaction is built and has multiple signatures
            tx.should.be.instanceof(lib_1.Transaction);
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            // Count non-empty signatures
            let signatureCount = 0;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    if (sig && sig.length > 0 && !sig.startsWith('0x' + '0'.repeat(130))) {
                        signatureCount++;
                    }
                }
            }
            signatureCount.should.be.greaterThanOrEqual(2, 'Should have at least 2 signatures for 2-of-2 multisig');
        });
    });
    describe('half-signing workflow with real transaction data', () => {
        // Import real multisig delegation transaction data from testnet
        const { HALF_SIGN_TEST_ACCOUNTS, MULTISIG_DELEGATION_PARAMS, MULTISIG_DELEGATION_UNSIGNED_TX_HEX, MULTISIG_DELEGATION_HALF_SIGNED_TX_HEX, MULTISIG_DELEGATION_FULLY_SIGNED_TX_HEX, } = require('../../resources/transactionData/multisigDelegationTx');
        it('should parse unsigned multisig delegation transaction from testnet', async () => {
            const rawTx = utils_1.default.removeHexPrefix(MULTISIG_DELEGATION_UNSIGNED_TX_HEX);
            // Verify transaction can be deserialized by FlareJS
            const decodedTxBytes = Buffer.from(rawTx, 'hex');
            decodedTxBytes.should.not.be.empty();
            decodedTxBytes.length.should.be.greaterThan(0);
        });
        it('should build delegation transaction with multisig addresses and verify sorting', async () => {
            const builder = factory.getDelegatorBuilder();
            const addresses = MULTISIG_DELEGATION_PARAMS.multisigAddresses;
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            builder
                .nodeID(MULTISIG_DELEGATION_PARAMS.nodeID)
                .startTime(MULTISIG_DELEGATION_PARAMS.startTime)
                .endTime(MULTISIG_DELEGATION_PARAMS.endTime)
                .stakeAmount(BigInt(MULTISIG_DELEGATION_PARAMS.stakeAmount))
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState({ capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' })
                .context(account_1.CONTEXT);
            const tx = await builder.build();
            // Verify transaction was built successfully
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify UTXOs are properly handled
            const txUtxos = tx._utxos;
            txUtxos.should.not.be.empty();
            txUtxos[0].should.have.property('addresses');
        });
        it('should successfully parse half-signed transaction', async () => {
            const rawTx = utils_1.default.removeHexPrefix(MULTISIG_DELEGATION_HALF_SIGNED_TX_HEX);
            const tx = (await factory.from(rawTx).build());
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify it has one signature
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            let signatureCount = 0;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    const sigHex = sig.toString('hex');
                    // Non-empty signature (not all zeros)
                    if (sigHex && !sigHex.match(/^0+$/)) {
                        signatureCount++;
                    }
                }
            }
            signatureCount.should.be.greaterThanOrEqual(1, 'Should have at least 1 signature in half-signed tx');
        });
        it('should successfully parse fully-signed transaction', async () => {
            const rawTx = utils_1.default.removeHexPrefix(MULTISIG_DELEGATION_FULLY_SIGNED_TX_HEX);
            const tx = (await factory.from(rawTx).build());
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify it has two signatures
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            let signatureCount = 0;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    const sigHex = sig.toString('hex');
                    // Non-empty signature (not all zeros)
                    if (sigHex && !sigHex.match(/^0+$/)) {
                        signatureCount++;
                    }
                }
            }
            signatureCount.should.be.greaterThanOrEqual(2, 'Should have at least 2 signatures in fully-signed tx');
        });
        it('should build, sign with user key, and match half-signed format', async () => {
            const builder = factory.getDelegatorBuilder();
            const addresses = MULTISIG_DELEGATION_PARAMS.multisigAddresses;
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            builder
                .nodeID(MULTISIG_DELEGATION_PARAMS.nodeID)
                .startTime(MULTISIG_DELEGATION_PARAMS.startTime)
                .endTime(MULTISIG_DELEGATION_PARAMS.endTime)
                .stakeAmount(BigInt(MULTISIG_DELEGATION_PARAMS.stakeAmount))
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState({ capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' })
                .context(account_1.CONTEXT)
                .sign({ key: HALF_SIGN_TEST_ACCOUNTS.user.privateKey });
            const tx = await builder.build();
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify transaction has credentials
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            // Count signatures
            let signatureCount = 0;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    const sigHex = sig.toString('hex');
                    if (sigHex && !sigHex.match(/^0+$/)) {
                        signatureCount++;
                    }
                }
            }
            signatureCount.should.be.greaterThanOrEqual(1, 'Should have user signature after signing');
        });
        it('should build and sign with both user and bitgo keys for fully-signed tx', async () => {
            const builder = factory.getDelegatorBuilder();
            const addresses = MULTISIG_DELEGATION_PARAMS.multisigAddresses;
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            builder
                .nodeID(MULTISIG_DELEGATION_PARAMS.nodeID)
                .startTime(MULTISIG_DELEGATION_PARAMS.startTime)
                .endTime(MULTISIG_DELEGATION_PARAMS.endTime)
                .stakeAmount(BigInt(MULTISIG_DELEGATION_PARAMS.stakeAmount))
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState({ capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' })
                .context(account_1.CONTEXT);
            // Sign with both keys
            builder.sign({ key: HALF_SIGN_TEST_ACCOUNTS.user.privateKey });
            builder.sign({ key: HALF_SIGN_TEST_ACCOUNTS.bitgo.privateKey });
            const tx = await builder.build();
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify transaction has both signatures
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            let signatureCount = 0;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    const sigHex = sig.toString('hex');
                    if (sigHex && !sigHex.match(/^0+$/)) {
                        signatureCount++;
                    }
                }
            }
            signatureCount.should.be.greaterThanOrEqual(2, 'Should have both user and bitgo signatures');
        });
        it('should build delegation transaction when addresses need sorting', async () => {
            // Create a scenario with addresses in different order
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            // Use addresses in a specific order
            const addresses = [account_1.ACCOUNT_3.address, account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + 14 * 24 * 60 * 60)
                .stakeAmount(BigInt(50000 * 1e9))
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState({ capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' })
                .context(account_1.CONTEXT);
            const tx = await builder.build();
            // Verify transaction builds successfully with sorted addresses
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify UTXOs are properly handled
            const txUtxos = tx._utxos;
            txUtxos.should.not.be.empty();
            txUtxos[0].should.have.property('addresses');
            txUtxos[0].addresses.length.should.equal(3);
        });
    });
    describe('recovery mode signing', () => {
        const mockFeeState = { capacity: 100000n, excess: 0n, price: 1000n, timestamp: '1234567890' };
        const MIN_DELEGATION_DURATION = 14 * 24 * 60 * 60;
        const MIN_DELEGATION_AMOUNT = BigInt(50000 * 1e9);
        /**
         * Recovery mode test verifies that when recoverSigner=true, the builder uses:
         * - Backup key (index 2) instead of user key (index 0)
         * - BitGo key (index 1) as always
         *
         * This is inherited from AtomicTransactionBuilder but should be verified
         * to work correctly with delegation transactions.
         */
        it('should sign delegation transaction in recovery mode (backup + bitgo keys)', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            // 2-of-3 multisig: user, bitgo, backup
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000', // 60k FLR
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            // Enable recovery mode - this tells the builder to use backup (index 2) instead of user (index 0)
            builder.recoverSigner = true;
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT);
            // In recovery mode: sign with backup (index 2) and bitgo (index 1)
            builder.sign({ key: account_1.ACCOUNT_3.privateKey }); // backup key
            builder.sign({ key: account_1.ACCOUNT_1.privateKey }); // bitgo key
            const tx = await builder.build();
            // Verify transaction builds successfully
            tx.should.be.instanceof(lib_1.Transaction);
            tx.type.should.equal(sdk_core_1.TransactionType.AddPermissionlessDelegator);
            // Verify credentials exist
            const credentials = tx._flareTransaction.credentials;
            credentials.should.not.be.empty();
            credentials.length.should.be.greaterThan(0);
            // Verify we have signatures from both recovery keys
            let signatureCount = 0;
            for (const credential of credentials) {
                const signatures = credential.getSignatures();
                for (const sig of signatures) {
                    const sigHex = sig.toString('hex');
                    // Non-empty signature (not all zeros)
                    if (sigHex && !sigHex.match(/^0+$/)) {
                        signatureCount++;
                    }
                }
            }
            // Should have at least 2 signatures (backup + bitgo)
            signatureCount.should.be.greaterThanOrEqual(2, 'Recovery mode should have backup + bitgo signatures');
        });
        it('should fail when signing in recovery mode with user key instead of backup', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            builder.recoverSigner = true;
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT);
            // Try to sign with user key (index 0) - this should fail in recovery mode
            // because recovery mode expects backup key (index 2) + bitgo key (index 1)
            builder.sign({ key: account_1.SEED_ACCOUNT.privateKey }); // user key (wrong for recovery mode)
            builder.sign({ key: account_1.ACCOUNT_1.privateKey }); // bitgo key
            const tx = await builder.build();
            // Transaction builds but may not have correct signature count
            // In recovery mode with wrong key, signature matching may fail
            tx.should.be.instanceof(lib_1.Transaction);
        });
        it('should verify recovery mode uses correct address indices (backup=2, bitgo=1)', async () => {
            const builder = factory.getDelegatorBuilder();
            const now = Math.floor(Date.now() / 1000);
            const validNodeID = 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL';
            const addresses = [account_1.SEED_ACCOUNT.addressTestnet, account_1.ACCOUNT_1.addressTestnet, account_1.ACCOUNT_3.address];
            const utxos = [
                {
                    outputID: 7,
                    txid: '2NWd9hrSGkWJWyTu4DnSM1qQSUT2DVm8uqwFk4wQRFLAmcsHQz',
                    outputidx: '0',
                    assetID: account_1.CONTEXT.avaxAssetID,
                    amount: '60000000000000',
                    threshold: 2,
                    addresses,
                    locktime: '0',
                },
            ];
            // Enable recovery mode
            builder.recoverSigner = true;
            builder
                .nodeID(validNodeID)
                .startTime(now + 60)
                .endTime(now + MIN_DELEGATION_DURATION)
                .stakeAmount(MIN_DELEGATION_AMOUNT)
                .fromPubKey(addresses)
                .decodedUtxos(utxos)
                .feeState(mockFeeState)
                .context(account_1.CONTEXT);
            // Build without signing to inspect internal state
            await builder.build();
            const transaction = builder.transaction;
            transaction._fromAddresses.length.should.equal(3);
            // In recovery mode, getSigningAddresses should return [backup, bitgo]
            const signingAddresses = builder.getSigningAddresses();
            signingAddresses.length.should.equal(2);
            // Verify signing addresses are backup (index 2) and bitgo (index 1)
            // Note: We can't directly compare addresses as they may be sorted,
            // but we can verify the method returns exactly 2 addresses
            signingAddresses[0].length.should.equal(20);
            signingAddresses[1].length.should.equal(20);
        });
    });
});
