// @bitgo/sdk-coin-avaxc/dist/test/unit/transactionBuilder/transfer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const sdk_coin_eth_1 = require("@bitgo/sdk-coin-eth");
const testData = __importStar(require("../../resources/avaxc"));
const sdk_core_1 = require("@bitgo/sdk-core");
const getBuilder_1 = require("../getBuilder");
describe('Avax C-Chain Transfer Transaction', function () {
    let txBuilder;
    const contractAddress = testData.TEST_ACCOUNT.ethAddress;
    const initTxBuilder = () => {
        txBuilder = (0, getBuilder_1.getBuilder)('tavaxc');
        txBuilder.fee({
            fee: '280000000000',
            gasLimit: '7000000',
        });
        txBuilder.counter(1);
        txBuilder.type(sdk_core_1.TransactionType.Send);
        txBuilder.contract(contractAddress);
    };
    it('Should build transfer tx', async function () {
        initTxBuilder();
        txBuilder
            .transfer()
            .amount('100000000000000000') // This represents 0.1 Avax = 0.1 Ether
            .contractSequenceId(1)
            .expirationTime(50000)
            .to(testData.TEST_ACCOUNT_2.ethAddress)
            .key(testData.OWNER_2.ethKey);
        txBuilder.sign({ key: testData.OWNER_1.ethKey });
        const tx = await txBuilder.build();
        const txJson = tx.toJson();
        should_1.default.exists(txJson.chainId);
        should_1.default.exists(txJson.from);
        should_1.default.exists(txJson.to);
        txJson.nonce.should.equals(1);
        txJson.from?.should.equals(testData.OWNER_1.ethAddress);
        txJson.chainId?.should.equals('0xa869');
        txJson.gasLimit.should.equals('7000000');
        txJson._type.should.equals(sdk_coin_eth_1.ETHTransactionType.LEGACY);
        txJson.gasPrice.should.equals('280000000000');
    });
    it('Should build with counter 0 if not manually defined', async function () {
        const builder = (0, getBuilder_1.getBuilder)('tavaxc');
        builder.fee({
            fee: '280000000000',
            gasLimit: '7000000',
        });
        builder.type(sdk_core_1.TransactionType.Send);
        builder.contract(testData.TEST_ACCOUNT.ethAddress);
        builder
            .transfer()
            .amount('100000000000000000') // This represents 0.1 Avax = 0.1 Ether
            .contractSequenceId(1)
            .expirationTime(50000)
            .to(testData.TEST_ACCOUNT_2.ethAddress)
            .key(testData.OWNER_2.ethKey);
        builder.sign({ key: testData.OWNER_1.ethKey });
        const tx = await builder.build();
        const txJson = tx.toJson();
        should_1.default.exists(txJson.chainId);
        should_1.default.exists(txJson.from);
        should_1.default.exists(txJson.to);
        txJson.nonce.should.equals(0);
        txJson.from.should.equals(testData.OWNER_1.ethAddress);
        txJson.chainId.should.equals('0xa869');
        txJson.gasLimit.should.equals('7000000');
        txJson._type.should.equals(sdk_coin_eth_1.ETHTransactionType.LEGACY);
        txJson.gasPrice.should.equals('280000000000');
    });
    it('Should build transfer with default type', async function () {
        const builder = (0, getBuilder_1.getBuilder)('tavaxc');
        builder.fee({
            fee: '280000000000',
            gasLimit: '7000000',
        });
        builder.counter(1);
        builder.contract(testData.TEST_ACCOUNT.ethAddress);
        builder
            .transfer()
            .amount('100000000000000000') // This represents 0.1 Avax = 0.1 Ether
            .contractSequenceId(1)
            .expirationTime(50000)
            .to(testData.TEST_ACCOUNT_2.ethAddress)
            .key(testData.OWNER_2.ethKey);
        builder.sign({ key: testData.OWNER_1.ethKey });
        const tx = await builder.build();
        const txJson = tx.toJson();
        should_1.default.exists(txJson.chainId);
        should_1.default.exists(txJson.from);
        should_1.default.exists(txJson.to);
        txJson.nonce.should.equals(1);
        txJson.from?.should.equals(testData.OWNER_1.ethAddress);
        txJson.chainId?.should.equals('0xa869');
        txJson.gasLimit.should.equals('7000000');
        txJson._type.should.equals(sdk_coin_eth_1.ETHTransactionType.LEGACY);
        txJson.gasPrice.should.equals('280000000000');
    });
    it('Should create transfer object if not created or return it if already initialized', async function () {
        const builder = (0, getBuilder_1.getBuilder)('tavaxc');
        builder.fee({
            fee: '280000000000',
            gasLimit: '7000000',
        });
        builder.counter(1);
        builder.contract(testData.TEST_ACCOUNT.ethAddress);
        builder.type(sdk_core_1.TransactionType.Send);
        const transferObj = builder.transfer().amount('5000').contractSequenceId(2).to(testData.TEST_ACCOUNT_2.ethAddress);
        const transferObj2 = builder.transfer();
        should_1.default.deepEqual(transferObj2, transferObj);
    });
    it('Should fail building transfer tx without fee', async function () {
        const builder = (0, getBuilder_1.getBuilder)('tavaxc');
        builder.counter(1);
        builder.type(sdk_core_1.TransactionType.Send);
        builder.contract(testData.TEST_ACCOUNT.ethAddress);
        builder
            .transfer()
            .amount('100000000000000000') // This represents 0.1 Avax = 0.1 Ether
            .contractSequenceId(1)
            .expirationTime(50000)
            .to(testData.TEST_ACCOUNT_2.ethAddress)
            .key(testData.OWNER_2.ethKey);
        builder.sign({ key: testData.OWNER_1.ethKey });
        builder.build().should.be.rejectedWith('Invalid transaction: missing fee');
    });
    it('Should fail getting transfer object with non-send type', async function () {
        const builder = (0, getBuilder_1.getBuilder)('tavaxc');
        builder.fee({
            fee: '280000000000',
            gasLimit: '7000000',
        });
        builder.counter(1);
        builder.contract(testData.TEST_ACCOUNT.ethAddress);
        builder.type(sdk_core_1.TransactionType.WalletInitialization);
        assert_1.default.throws(() => builder.transfer(), (e) => e.message === 'Transfers can only be set for send transactions');
    });
    it('Should build unsigned transfer with final v', async function () {
        initTxBuilder();
        txBuilder
            .transfer()
            .amount('100000000000000000') // This represents 0.1 Avax = 0.1 Ether
            .contractSequenceId(1)
            .expirationTime(50000)
            .to(testData.TEST_ACCOUNT_2.ethAddress)
            .key(testData.OWNER_2.ethKey);
        const tx = await txBuilder.build();
        const txJson = tx.toJson();
        should_1.default.exists(txJson.chainId);
        should_1.default.exists(txJson.to);
        txJson.nonce.should.equals(1);
        should_1.default.equal(txJson.chainId, '0xa869');
        txJson.gasLimit.should.equals('7000000');
        should_1.default.equal(txJson.gasPrice, '280000000000');
        should_1.default.equal(txJson.v, '0x0150f5');
    });
    it('Should build fee market transaction', async function () {
        initTxBuilder();
        txBuilder.fee({
            fee: '280000000000',
            eip1559: {
                maxPriorityFeePerGas: '5',
                maxFeePerGas: '30',
            },
            gasLimit: '7000000',
        });
        txBuilder
            .transfer()
            .amount('100000000000000000') // This represents 0.1 Avax = 0.1 Ether
            .contractSequenceId(1)
            .expirationTime(50000)
            .to(testData.TEST_ACCOUNT_2.ethAddress)
            .key(testData.OWNER_2.ethKey);
        const tx = await txBuilder.build();
        const txJson = tx.toJson();
        txJson._type.should.equals(sdk_coin_eth_1.ETHTransactionType.EIP1559);
        should_1.default.exists(txJson.chainId);
        should_1.default.exists(txJson.to);
        txJson.nonce.should.equals(1);
        should_1.default.equal(txJson.chainId, '0xa869');
        txJson.gasLimit.should.equals('7000000');
        should_1.default.equal(txJson.maxPriorityFeePerGas, '5');
        should_1.default.equal(txJson.maxFeePerGas, '30');
    });
});
