// @bitgo/sdk-coin-ton/dist/test/unit/tokenTransferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const tonweb_1 = __importDefault(require("tonweb"));
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const src_1 = require("../../src");
const testData = __importStar(require("../resources/ton"));
const utils = __importStar(require("../../src/lib/utils"));
describe('Ton Token Transfer Builder', () => {
    const tokenName = 'tton:ukwny-us';
    const factory = new src_1.TransactionBuilderFactory(statics_1.coins.get(tokenName));
    it('should build a unsigned transfer tx', async function () {
        const txId = testData.signedTokenSendTransaction.txId.replace(/\//g, '_').replace(/\+/g, '-');
        const txBuilder = factory.getTokenTransferBuilder();
        txBuilder.sender(testData.tokenSender.address);
        txBuilder.sequenceNumber(0);
        txBuilder.publicKey(testData.tokenSender.publicKey);
        txBuilder.expireTime(testData.tokenSender.expireTime);
        txBuilder.send(testData.tokenRecipients[0]);
        txBuilder.setSenderJettonWalletAddress(testData.tokenSender.jettonAddress);
        txBuilder.setTonAmount(testData.tokenSender.tonAmount); // amount of TON to send to the sender's jetton wallet
        txBuilder.setForwardTonAmount(testData.tokenSender.forwardTonAmount); // amount of TON to cover forward fees in case of notify transfer
        txBuilder.setMessage(testData.tokenSender.message);
        const tx = await txBuilder.build();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.SendToken);
        should_1.default.equal(tx.toJson().bounceable, false);
        tx.inputs.length.should.equal(1);
        tx.inputs[0].should.deepEqual({
            address: testData.tokenSender.address,
            value: testData.tokenRecipients[0].amount,
            coin: tokenName,
        });
        tx.outputs.length.should.equal(1);
        tx.outputs[0].should.deepEqual({
            address: testData.tokenRecipients[0].address,
            value: testData.tokenRecipients[0].amount,
            coin: tokenName,
        });
        tx.id.should.equal(txId);
        const rawTx = tx.toBroadcastFormat();
        rawTx.should.equal(testData.signedTokenSendTransaction.tx);
    });
    it('should build a unsigned transfer tx with bounceable flag', async function () {
        const txId = testData.signedTokenSendTransaction.txIdBounceable.replace(/\//g, '_').replace(/\+/g, '-');
        const txBuilder = factory.getTokenTransferBuilder();
        txBuilder.sender(testData.tokenSender.address);
        txBuilder.sequenceNumber(0);
        txBuilder.publicKey(testData.tokenSender.publicKey);
        txBuilder.expireTime(testData.tokenSender.expireTime);
        txBuilder.send(testData.tokenRecipients[0]);
        txBuilder.setSenderJettonWalletAddress(testData.tokenSender.jettonAddress);
        txBuilder.setTonAmount(testData.tokenSender.tonAmount); // amount of TON to send to the sender's jetton wallet
        txBuilder.setForwardTonAmount(testData.tokenSender.forwardTonAmount); // amount of TON to cover forward fees in case of notify transfer
        txBuilder.setMessage(testData.tokenSender.message);
        txBuilder.bounceable(true);
        const tx = await txBuilder.build();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.SendToken);
        should_1.default.equal(tx.toJson().bounceable, true);
        tx.inputs.length.should.equal(1);
        tx.inputs[0].should.deepEqual({
            address: testData.tokenSender.address,
            value: testData.tokenRecipients[0].amount,
            coin: tokenName,
        });
        tx.outputs.length.should.equal(1);
        tx.outputs[0].should.deepEqual({
            address: testData.tokenRecipients[0].address,
            value: testData.tokenRecipients[0].amount,
            coin: tokenName,
        });
        tx.id.should.equal(txId);
        const rawTx = tx.toBroadcastFormat();
        rawTx.should.equal(testData.signedTokenSendTransaction.txBounceable);
    });
    it('should build a send from rawTx', async function () {
        const txBuilder = factory.from(testData.signedTokenSendTransaction.tx);
        const builtTx = await txBuilder.build();
        const jsonTx = builtTx.toJson();
        should_1.default.equal(builtTx.type, sdk_core_1.TransactionType.SendToken);
        should_1.default.equal(builtTx.signablePayload.toString('base64'), testData.signedTokenSendTransaction.signable);
        should_1.default.equal(builtTx.id, testData.signedTokenSendTransaction.txId);
        const builder2 = factory.from(builtTx.toBroadcastFormat());
        const builtTx2 = await builder2.build();
        should_1.default.equal(builtTx2.type, sdk_core_1.TransactionType.SendToken);
        should_1.default.equal(builtTx.toBroadcastFormat(), testData.signedTokenSendTransaction.tx);
        builtTx.inputs.length.should.equal(1);
        builtTx.outputs.length.should.equal(1);
        jsonTx.sender.should.equal('EQCqQzfyg0cZ-8t9v2YoHmFFxG5jgjvQRTZ2yeDjO5z5ZRy9');
        jsonTx.destination.should.equal('EQB-CM6DF-jpq9XVdiSdefAMU5KC1gpZuYBFp-Q65aUhnx5K');
        jsonTx.amount.should.equal(testData.tokenRecipients[0].amount);
        jsonTx.seqno.should.equal(0);
        jsonTx.expirationTime.should.equal(testData.tokenSender.expireTime);
        const builtTx3 = await txBuilder.bounceable(true).fromAddressBounceable(false).toAddressBounceable(false).build();
        txBuilder.from(testData.signedTokenSendTransaction.txBounceable);
        const jsonTx3 = builtTx3.toJson();
        should_1.default.equal(jsonTx3.bounceable, true);
        should_1.default.equal(builtTx3.signablePayload.toString('base64'), testData.signedTokenSendTransaction.bounceableSignable);
        should_1.default.equal(builtTx3.id, testData.signedTokenSendTransaction.txIdBounceable);
        should_1.default.equal(builtTx3.toBroadcastFormat(), testData.signedTokenSendTransaction.txBounceable);
        jsonTx3.sender.should.equal(testData.tokenSender.address);
        jsonTx3.destination.should.equal(testData.tokenRecipients[0].address);
        jsonTx3.amount.should.equal(testData.tokenRecipients[0].amount);
        jsonTx3.seqno.should.equal(0);
        jsonTx3.expirationTime.should.equal(testData.tokenSender.expireTime);
    });
    it('should parse a raw transaction and set flags', async function () {
        const factory = new src_1.TransactionBuilderFactory(statics_1.coins.get(tokenName));
        const txBuilder = factory.from(testData.signedTokenSendTransaction.tx);
        const txBuilderBounceable = factory.from(testData.signedTokenSendTransaction.txBounceable);
        const tx = await txBuilder.build();
        const txBounceable = await txBuilderBounceable.build();
        tx.toJson().bounceable.should.equal(false);
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.SendToken);
        txBounceable.toJson().bounceable.should.equal(true);
        should_1.default.equal(txBounceable.type, sdk_core_1.TransactionType.SendToken);
    });
    xit('should build a signed transfer tx and submit onchain', async function () {
        const tonweb = new tonweb_1.default(new tonweb_1.default.HttpProvider('https://testnet.toncenter.com/api/v2/jsonRPC'));
        const keyPair = new src_1.KeyPair({ prv: testData.privateKeys.prvKey6 });
        const publicKey = keyPair.getKeys().pub;
        const address = await utils.default.getAddressFromPublicKey(publicKey);
        const txBuilder = factory.getTokenTransferBuilder();
        txBuilder.sender(address);
        const WalletClass = tonweb.wallet.all['v4R2'];
        const wallet = new WalletClass(tonweb.provider, {
            publicKey: tonweb.utils.hexToBytes(publicKey),
            wc: 0,
        });
        const seqno = await wallet.methods.seqno().call();
        txBuilder.sequenceNumber(seqno);
        txBuilder.publicKey(publicKey);
        const expireAt = Math.floor(Date.now() / 1e3) + 60 * 60 * 24 * 7; // 7 days
        txBuilder.expireTime(expireAt);
        txBuilder.send({
            address: testData.tokenRecipients[0].address,
            amount: '1000000000',
        });
        txBuilder.setSenderJettonWalletAddress(testData.tokenSender.jettonAddress);
        txBuilder.setTonAmount('500000000'); // amount of TON to send to the sender's jetton wallet
        txBuilder.setForwardTonAmount('100000000'); // amount of TON to cover forward fees in case of notify transfer
        txBuilder.setMessage('test');
        const tx = await txBuilder.build();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.SendToken);
        const signable = tx.signablePayload;
        const signature = keyPair.signMessageinUint8Array(signable);
        const signedTx = await txBuilder.build();
        const builder2 = factory.from(signedTx.toBroadcastFormat());
        builder2.addSignature(keyPair.getKeys(), Buffer.from(signature));
        const tx2 = await builder2.build();
        const signature2 = keyPair.signMessageinUint8Array(tx2.signablePayload);
        should_1.default.equal(Buffer.from(signature).toString('hex'), Buffer.from(signature2).toString('hex'));
        await new Promise((resolve) => setTimeout(resolve, 2000));
        try {
            const result = await tonweb.provider.sendBoc(tx2.toBroadcastFormat());
            console.log('Network Response:', result);
        }
        catch (error) {
            console.error('Error Response:', error);
        }
    });
    it('should build a signed transfer tx using add signature', async function () {
        const keyPair = new src_1.KeyPair({ prv: testData.privateKeys.prvKey6 });
        const publicKey = keyPair.getKeys().pub;
        const address = await utils.default.getAddressFromPublicKey(publicKey);
        const txBuilder = factory.getTokenTransferBuilder();
        txBuilder.sender(address);
        txBuilder.sequenceNumber(0);
        txBuilder.publicKey(publicKey);
        const expireAt = Math.floor(Date.now() / 1e3) + 60 * 60 * 24 * 7;
        txBuilder.expireTime(expireAt);
        txBuilder.send(testData.tokenRecipients[0]);
        txBuilder.setSenderJettonWalletAddress(testData.tokenSender.jettonAddress);
        txBuilder.setTonAmount(testData.tokenSender.tonAmount); // amount of TON to send to the sender's jetton wallet
        txBuilder.setForwardTonAmount(testData.tokenSender.forwardTonAmount); // amount of TON to cover forward fees in case of notify transfer
        txBuilder.setMessage(testData.tokenSender.message);
        const tx = await txBuilder.build();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.SendToken);
        const signable = tx.signablePayload;
        const signature = keyPair.signMessageinUint8Array(signable);
        txBuilder.addSignature(keyPair.getKeys(), Buffer.from(signature));
        const signedTx = await txBuilder.build();
        const builder2 = factory.from(signedTx.toBroadcastFormat());
        const tx2 = await builder2.build();
        const signature2 = keyPair.signMessageinUint8Array(tx2.signablePayload);
        should_1.default.equal(Buffer.from(signature).toString('hex'), Buffer.from(signature2).toString('hex'));
        should_1.default.equal(tx.toBroadcastFormat(), tx2.toBroadcastFormat());
    });
    it('should build transfer tx for non-bounceable address', async function () {
        const txBuilder = factory.getTokenTransferBuilder();
        txBuilder.sender(testData.tokenSender.address);
        txBuilder.sequenceNumber(0);
        txBuilder.publicKey(testData.tokenSender.publicKey);
        txBuilder.expireTime(testData.tokenSender.expireTime);
        const address = 'EQAWzEKcdnykvXfUNouqdS62tvrp32bCxuKS6eQrS6ISgcLo';
        const otherFormat = 'UQAWzEKcdnykvXfUNouqdS62tvrp32bCxuKS6eQrS6ISgZ8t';
        const amount = testData.tokenRecipients[0].amount;
        txBuilder.send({ address, amount });
        txBuilder.setSenderJettonWalletAddress(testData.tokenSender.jettonAddress);
        txBuilder.setTonAmount(testData.tokenSender.tonAmount); // amount of TON to send to the sender's jetton wallet
        txBuilder.setForwardTonAmount(testData.tokenSender.forwardTonAmount); // amount of TON to cover forward fees in case of notify transfer
        txBuilder.setMessage(testData.tokenSender.message);
        const tx = await txBuilder.build();
        should_1.default.equal(tx.type, sdk_core_1.TransactionType.SendToken);
        tx.inputs.length.should.equal(1);
        tx.inputs[0].should.deepEqual({
            address: testData.tokenSender.address,
            value: amount,
            coin: tokenName,
        });
        tx.outputs.length.should.equal(1);
        tx.outputs[0].should.deepEqual({
            address,
            value: amount,
            coin: tokenName,
        });
        const txJson = tx.toJson();
        txJson.destination.should.equal(address);
        txJson.destinationAlias.should.equal(otherFormat);
        const builder2 = factory.from(tx.toBroadcastFormat());
        const tx2 = await builder2.build();
        const txJson2 = tx2.toJson();
        txJson2.destinationAlias.should.equal(otherFormat);
    });
});
