// @bitgo/wasm-utxo/dist/cjs/js/descriptorWallet/psbt/createPsbt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_BIP125_RBF_SEQUENCE = void 0;
exports.findTapLeafScript = findTapLeafScript;
exports.createPsbt = createPsbt;
/**
 * PSBT creation for descriptor wallets.
 * Moved from @bitgo/utxo-core.
 *
 * This version uses wasm-utxo Psbt directly without the wrap/unwrap pattern.
 */
const index_js_1 = require("../../index.js");
const assertSatisfiable_js_1 = require("./assertSatisfiable.js");
function bytesEqual(a, b) {
    if (a.length !== b.length)
        return false;
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i])
            return false;
    }
    return true;
}
function toHex(bytes) {
    return Array.from(bytes)
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
}
/**
 * Non-Final (Replaceable)
 * Reference: https://github.com/bitcoin/bitcoin/blob/v25.1/src/rpc/rawtransaction_util.cpp#L49
 * */
exports.MAX_BIP125_RBF_SEQUENCE = 0xffffffff - 2;
function findTapLeafScript(input, script) {
    if (!(script instanceof Uint8Array)) {
        script = script.encode();
    }
    const scriptBytes = script;
    const matches = input.filter((leaf) => {
        return bytesEqual(leaf.script, scriptBytes);
    });
    if (matches.length === 0) {
        throw new Error(`No tapLeafScript found for script: ${toHex(scriptBytes)}`);
    }
    if (matches.length > 1) {
        throw new Error(`Multiple tapLeafScripts found for script: ${toHex(scriptBytes)}`);
    }
    return matches[0];
}
/**
 * Create a PSBT for descriptor wallet transactions.
 *
 * This function uses wasm-utxo Psbt directly without any wrap/unwrap conversions.
 *
 * @param params - PSBT parameters (version, locktime, sequence)
 * @param inputs - Descriptor wallet inputs
 * @param outputs - Outputs with optional descriptors for change
 * @returns A wasm-utxo Psbt instance
 */
function createPsbt(params, inputs, outputs) {
    const psbt = new index_js_1.Psbt(params.version ?? 2, params.locktime ?? 0);
    // Add inputs
    for (const input of inputs) {
        const sequence = input.sequence ?? params.sequence ?? exports.MAX_BIP125_RBF_SEQUENCE;
        psbt.addInput(input.hash, input.index, input.witnessUtxo.value, input.witnessUtxo.script, sequence);
    }
    // Add outputs
    for (const output of outputs) {
        psbt.addOutput(output.script, output.value);
    }
    // Update inputs with descriptor metadata
    for (const [inputIndex, input] of inputs.entries()) {
        (0, assertSatisfiable_js_1.assertSatisfiable)(psbt, inputIndex, input.descriptor);
        psbt.updateInputWithDescriptor(inputIndex, input.descriptor);
    }
    // Update outputs with descriptor metadata (for change outputs)
    for (const [outputIndex, output] of outputs.entries()) {
        if (output.descriptor) {
            psbt.updateOutputWithDescriptor(outputIndex, output.descriptor);
        }
    }
    return psbt;
}
