// @bitgo/account-lib/dist/test/unit/coin/baseCoin/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sinon_1 = __importDefault(require("sinon"));
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const testTransactionBuilder_1 = require("../../../resources/testTransactionBuilder");
const testTransaction_1 = require("../../../resources/testTransaction");
describe('Transaction builder', () => {
    let txBuilder;
    let sandbox;
    beforeEach(() => {
        sandbox = sinon_1.default.createSandbox();
        txBuilder = new testTransactionBuilder_1.TestTransactionBuilder();
    });
    afterEach(() => {
        sandbox.restore();
    });
    it('should sign a transaction that is valid', () => {
        const testTx = sinon_1.default.createStubInstance(testTransaction_1.TestTransaction);
        testTx.canSign.returns(true);
        const validateKey = sinon_1.default.spy(txBuilder, 'validateKey');
        txBuilder.from(testTx);
        txBuilder.sign({ key: 'validKey' });
        sandbox.assert.calledOnce(validateKey);
    });
    it('should sign a transaction with an invalid signature', () => {
        const testTx = sinon_1.default.createStubInstance(testTransaction_1.TestTransaction);
        testTx.canSign.returns(false);
        const validateKey = sinon_1.default.spy(txBuilder, 'validateKey');
        txBuilder.from(testTx);
        assert_1.default.throws(() => txBuilder.sign({ key: 'invalidKey' }));
        sandbox.assert.calledOnce(validateKey);
    });
    it('should parse a valid transaction', () => {
        const testTx = sinon_1.default.createStubInstance(testTransaction_1.TestTransaction);
        testTx.canSign.returns(true);
        const validateRawTransaction = sinon_1.default.spy(txBuilder, 'validateRawTransaction');
        const fromImplementation = sinon_1.default.spy(txBuilder, 'fromImplementation');
        txBuilder.from(testTx);
        sandbox.assert.calledOnce(validateRawTransaction);
        sandbox.assert.calledOnce(fromImplementation);
    });
    it('should build a valid transaction', () => {
        const testTx = sinon_1.default.createStubInstance(testTransaction_1.TestTransaction);
        testTx.canSign.returns(true);
        const validateTransaction = sinon_1.default.spy(txBuilder, 'validateTransaction');
        const buildImplementation = sinon_1.default.spy(txBuilder, 'buildImplementation');
        txBuilder.from(testTx);
        txBuilder.build();
        sandbox.assert.calledOnce(validateTransaction);
        sandbox.assert.calledOnce(buildImplementation);
    });
    it('should verify validity window params', () => {
        const testTx = sinon_1.default.createStubInstance(testTransaction_1.TestTransaction);
        testTx.canSign.returns(true);
        txBuilder.from(testTx);
        txBuilder.build();
        let validityWindow;
        let params = {};
        (0, should_1.default)(() => txBuilder.getValidityWindow(params)).throwError('Unit parameter must be specified as blockheight or timestamp');
        params = {
            unit: 'wrongUnit',
        };
        (0, should_1.default)(() => txBuilder.getValidityWindow(params)).throwError('Unit parameter must be specified as blockheight or timestamp');
        params = {
            firstValid: 10,
            lastValid: 23,
            minDuration: 10,
            maxDuration: 20,
            unit: '',
        };
        (0, should_1.default)(() => txBuilder.getValidityWindow(params)).throwError('Unit parameter must be specified as blockheight or timestamp');
        params = {
            firstValid: 10,
            lastValid: 23,
            minDuration: 10,
            maxDuration: 5,
            unit: 'blockheight',
        };
        (0, should_1.default)(() => txBuilder.getValidityWindow(params)).throwError(`Expected maxDuration (5) to be grather than minDuration (10)`);
        params = {
            firstValid: 10,
            lastValid: 11,
            minDuration: 10,
            maxDuration: 20,
            unit: 'timestamp',
        };
        validityWindow = txBuilder.getValidityWindow(params);
        validityWindow.should.have.properties(['firstValid', 'lastValid', 'minDuration', 'maxDuration', 'unit']);
        validityWindow.firstValid.should.be.equal(10);
        validityWindow.lastValid.should.be.equal(30);
        validityWindow.minDuration.should.be.equal(10);
        validityWindow.maxDuration.should.be.equal(20);
        validityWindow.unit.should.be.equal('timestamp');
        params = {
            firstValid: 10,
            lastValid: 23,
            minDuration: 10,
            maxDuration: 11,
            unit: 'blockheight',
        };
        validityWindow = txBuilder.getValidityWindow(params);
        validityWindow.should.have.properties(['firstValid', 'lastValid', 'minDuration', 'maxDuration', 'unit']);
        validityWindow.firstValid.should.be.equal(10);
        validityWindow.lastValid.should.be.equal(21);
        validityWindow.minDuration.should.be.equal(10);
        validityWindow.maxDuration.should.be.equal(11);
        validityWindow.unit.should.be.equal('blockheight');
        params = {
            unit: 'timestamp',
        };
        const dateNow = Date.now();
        validityWindow = txBuilder.getValidityWindow(params);
        validityWindow.should.have.properties(['firstValid', 'lastValid', 'minDuration', 'maxDuration', 'unit']);
        validityWindow.firstValid.should.be.greaterThanOrEqual(dateNow);
        validityWindow.lastValid.should.be.equal(validityWindow.firstValid + 31536000000);
        validityWindow.minDuration.should.be.equal(0);
        validityWindow.maxDuration.should.be.equal(31536000000);
        validityWindow.unit.should.be.equal('timestamp');
        params = {
            unit: 'blockheight',
        };
        validityWindow = txBuilder.getValidityWindow(params);
        validityWindow.should.have.properties(['firstValid', 'lastValid', 'minDuration', 'maxDuration', 'unit']);
        validityWindow.firstValid.should.be.equal(0);
        validityWindow.lastValid.should.be.equal(1000000);
        validityWindow.minDuration.should.be.equal(0);
        validityWindow.maxDuration.should.be.equal(1000000);
        validityWindow.unit.should.be.equal('blockheight');
    });
});
