// @bitgo/account-lib/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apt = exports.Coredao = exports.Oas = exports.zkEth = exports.XRP = exports.Ton = exports.Sui = exports.Opeth = exports.Polyx = exports.Polygon = exports.Bsc = exports.Near = exports.Dot = exports.Ada = exports.Sol = exports.Asi = exports.Initia = exports.Cronos = exports.Mon = exports.Baby = exports.Rune = exports.Coreum = exports.Zeta = exports.Islm = exports.Injective = exports.Sei = exports.Hash = exports.Bld = exports.Bera = exports.Tia = exports.AvaxP = exports.Osmo = exports.Atom = exports.Arbeth = exports.Algo = exports.Stx = exports.Cspr = exports.Hbar = exports.Celo = exports.Rbtc = exports.AvaxC = exports.Etc = exports.Eth = exports.Xtz = exports.Trx = exports.BaseCoin = exports.Ed25519KeyDeriver = exports.crypto = exports.Eddsa = exports.Ed25519BIP32 = void 0;
exports.supportedCoins = exports.MIDNIGHT_TNC_HASH = exports.FlrP = exports.Canton = exports.CosmosSharedCoin = exports.Vet = exports.Soneium = exports.Stt = exports.World = exports.Wemix = exports.Xdc = exports.Sgb = exports.EvmCoin = exports.Flr = exports.Icp = exports.Tempo = exports.Tao = exports.Apechain = void 0;
exports.getBuilder = getBuilder;
exports.getMessageBuilderFactory = getMessageBuilderFactory;
exports.register = register;
exports.registerMessageBuilderFactory = registerMessageBuilderFactory;
exports.verifyMessage = verifyMessage;
const sdk_core_1 = require("@bitgo/sdk-core");
Object.defineProperty(exports, "Ed25519KeyDeriver", { enumerable: true, get: function () { return sdk_core_1.Ed25519KeyDeriver; } });
Object.defineProperty(exports, "Ed25519BIP32", { enumerable: true, get: function () { return sdk_core_1.Ed25519BIP32; } });
Object.defineProperty(exports, "Eddsa", { enumerable: true, get: function () { return sdk_core_1.Eddsa; } });
const statics_1 = require("@bitgo/statics");
/**
 * Deprecated after version 2.19.0
 * Retained for short term backwards compatibility - migrate to: @bitgo/sdk-api and @bitgo/sdk-core
 */
exports.crypto = sdk_core_1.acountLibCrypto;
/**
 * Deprecated after version 2.19.0
 * Retained for short term backwards compatibility - migrate to: @bitgo/sdk-api and @bitgo/sdk-core
 */
exports.BaseCoin = sdk_core_1.accountLibBaseCoin;
// coins
const Trx = __importStar(require("@bitgo/sdk-coin-trx"));
exports.Trx = Trx;
const sdk_coin_xtz_1 = require("@bitgo/sdk-coin-xtz");
Object.defineProperty(exports, "Xtz", { enumerable: true, get: function () { return sdk_coin_xtz_1.XtzLib; } });
const Eth = __importStar(require("@bitgo/sdk-coin-eth"));
exports.Eth = Eth;
const Etc = __importStar(require("@bitgo/sdk-coin-etc"));
exports.Etc = Etc;
const AvaxC = __importStar(require("@bitgo/sdk-coin-avaxc"));
exports.AvaxC = AvaxC;
const Rbtc = __importStar(require("@bitgo/sdk-coin-rbtc"));
exports.Rbtc = Rbtc;
const Celo = __importStar(require("@bitgo/sdk-coin-celo"));
exports.Celo = Celo;
const Hbar = __importStar(require("@bitgo/sdk-coin-hbar"));
exports.Hbar = Hbar;
const sdk_coin_cspr_1 = require("@bitgo/sdk-coin-cspr");
Object.defineProperty(exports, "Cspr", { enumerable: true, get: function () { return sdk_coin_cspr_1.CsprLib; } });
const sdk_coin_stx_1 = require("@bitgo/sdk-coin-stx");
Object.defineProperty(exports, "Stx", { enumerable: true, get: function () { return sdk_coin_stx_1.StxLib; } });
const sdk_coin_algo_1 = require("@bitgo/sdk-coin-algo");
Object.defineProperty(exports, "Algo", { enumerable: true, get: function () { return sdk_coin_algo_1.AlgoLib; } });
const Arbeth = __importStar(require("@bitgo/sdk-coin-arbeth"));
exports.Arbeth = Arbeth;
const Atom = __importStar(require("@bitgo/sdk-coin-atom"));
exports.Atom = Atom;
const Osmo = __importStar(require("@bitgo/sdk-coin-osmo"));
exports.Osmo = Osmo;
const sdk_coin_avaxp_1 = require("@bitgo/sdk-coin-avaxp");
Object.defineProperty(exports, "AvaxP", { enumerable: true, get: function () { return sdk_coin_avaxp_1.AvaxpLib; } });
const Tia = __importStar(require("@bitgo/sdk-coin-tia"));
exports.Tia = Tia;
const Bera = __importStar(require("@bitgo/sdk-coin-bera"));
exports.Bera = Bera;
const Bld = __importStar(require("@bitgo/sdk-coin-bld"));
exports.Bld = Bld;
const Hash = __importStar(require("@bitgo/sdk-coin-hash"));
exports.Hash = Hash;
const Sei = __importStar(require("@bitgo/sdk-coin-sei"));
exports.Sei = Sei;
const Injective = __importStar(require("@bitgo/sdk-coin-injective"));
exports.Injective = Injective;
const Islm = __importStar(require("@bitgo/sdk-coin-islm"));
exports.Islm = Islm;
const Zeta = __importStar(require("@bitgo/sdk-coin-zeta"));
exports.Zeta = Zeta;
const Coreum = __importStar(require("@bitgo/sdk-coin-coreum"));
exports.Coreum = Coreum;
const Rune = __importStar(require("@bitgo/sdk-coin-rune"));
exports.Rune = Rune;
const Baby = __importStar(require("@bitgo/sdk-coin-baby"));
exports.Baby = Baby;
const Mon = __importStar(require("@bitgo/sdk-coin-mon"));
exports.Mon = Mon;
const Cronos = __importStar(require("@bitgo/sdk-coin-cronos"));
exports.Cronos = Cronos;
const Initia = __importStar(require("@bitgo/sdk-coin-initia"));
exports.Initia = Initia;
const Asi = __importStar(require("@bitgo/sdk-coin-asi"));
exports.Asi = Asi;
const Sol = __importStar(require("@bitgo/sdk-coin-sol"));
exports.Sol = Sol;
const Ada = __importStar(require("@bitgo/sdk-coin-ada"));
exports.Ada = Ada;
const Dot = __importStar(require("@bitgo/sdk-coin-dot"));
exports.Dot = Dot;
const Near = __importStar(require("@bitgo/sdk-coin-near"));
exports.Near = Near;
const Bsc = __importStar(require("@bitgo/sdk-coin-bsc"));
exports.Bsc = Bsc;
const Polygon = __importStar(require("@bitgo/sdk-coin-polygon"));
exports.Polygon = Polygon;
const Polyx = __importStar(require("@bitgo/sdk-coin-polyx"));
exports.Polyx = Polyx;
const Opeth = __importStar(require("@bitgo/sdk-coin-opeth"));
exports.Opeth = Opeth;
const Sui = __importStar(require("@bitgo/sdk-coin-sui"));
exports.Sui = Sui;
const Ton = __importStar(require("@bitgo/sdk-coin-ton"));
exports.Ton = Ton;
const XRP = __importStar(require("@bitgo/sdk-coin-xrp"));
exports.XRP = XRP;
const zkEth = __importStar(require("@bitgo/sdk-coin-zketh"));
exports.zkEth = zkEth;
const Oas = __importStar(require("@bitgo/sdk-coin-oas"));
exports.Oas = Oas;
const Coredao = __importStar(require("@bitgo/sdk-coin-coredao"));
exports.Coredao = Coredao;
const Apt = __importStar(require("@bitgo/sdk-coin-apt"));
exports.Apt = Apt;
const Apechain = __importStar(require("@bitgo/sdk-coin-apechain"));
exports.Apechain = Apechain;
const Tao = __importStar(require("@bitgo/sdk-coin-tao"));
exports.Tao = Tao;
const Tempo = __importStar(require("@bitgo/sdk-coin-tempo"));
exports.Tempo = Tempo;
const Icp = __importStar(require("@bitgo/sdk-coin-icp"));
exports.Icp = Icp;
const Flr = __importStar(require("@bitgo/sdk-coin-flr"));
exports.Flr = Flr;
const EvmCoin = __importStar(require("@bitgo/sdk-coin-evm"));
exports.EvmCoin = EvmCoin;
const Sgb = __importStar(require("@bitgo/sdk-coin-sgb"));
exports.Sgb = Sgb;
const Xdc = __importStar(require("@bitgo/sdk-coin-xdc"));
exports.Xdc = Xdc;
const Wemix = __importStar(require("@bitgo/sdk-coin-wemix"));
exports.Wemix = Wemix;
const World = __importStar(require("@bitgo/sdk-coin-world"));
exports.World = World;
const Stt = __importStar(require("@bitgo/sdk-coin-stt"));
exports.Stt = Stt;
const Soneium = __importStar(require("@bitgo/sdk-coin-soneium"));
exports.Soneium = Soneium;
const Vet = __importStar(require("@bitgo/sdk-coin-vet"));
exports.Vet = Vet;
const CosmosSharedCoin = __importStar(require("@bitgo/sdk-coin-cosmos"));
exports.CosmosSharedCoin = CosmosSharedCoin;
const Canton = __importStar(require("@bitgo/sdk-coin-canton"));
exports.Canton = Canton;
const sdk_coin_flrp_1 = require("@bitgo/sdk-coin-flrp");
Object.defineProperty(exports, "FlrP", { enumerable: true, get: function () { return sdk_coin_flrp_1.FlrPLib; } });
const utils_1 = require("./utils");
Object.defineProperty(exports, "MIDNIGHT_TNC_HASH", { enumerable: true, get: function () { return utils_1.MIDNIGHT_TNC_HASH; } });
const coinBuilderMap = {
    trx: Trx.WrappedBuilder,
    ttrx: Trx.WrappedBuilder,
    xtz: sdk_coin_xtz_1.XtzLib.TransactionBuilder,
    txtz: sdk_coin_xtz_1.XtzLib.TransactionBuilder,
    etc: Etc.TransactionBuilder,
    tetc: Etc.TransactionBuilder,
    eth: Eth.TransactionBuilder,
    teth: Eth.TransactionBuilder,
    gteth: Eth.TransactionBuilder,
    hteth: Eth.TransactionBuilder,
    rbtc: Rbtc.TransactionBuilder,
    trbtc: Rbtc.TransactionBuilder,
    celo: Celo.TransactionBuilder,
    tcelo: Celo.TestnetTransactionBuilder,
    avaxc: AvaxC.TransactionBuilder,
    tavaxc: AvaxC.TransactionBuilder,
    bsc: Bsc.TransactionBuilder,
    tbsc: Bsc.TransactionBuilder,
    avaxp: sdk_coin_avaxp_1.AvaxpLib.TransactionBuilderFactory,
    tavaxp: sdk_coin_avaxp_1.AvaxpLib.TransactionBuilderFactory,
    hbar: Hbar.TransactionBuilderFactory,
    thbar: Hbar.TransactionBuilderFactory,
    cspr: sdk_coin_cspr_1.CsprLib.TransactionBuilderFactory,
    tcspr: sdk_coin_cspr_1.CsprLib.TransactionBuilderFactory,
    stx: sdk_coin_stx_1.StxLib.TransactionBuilderFactory,
    tstx: sdk_coin_stx_1.StxLib.TransactionBuilderFactory,
    algo: sdk_coin_algo_1.AlgoLib.TransactionBuilderFactory,
    talgo: sdk_coin_algo_1.AlgoLib.TransactionBuilderFactory,
    sol: Sol.TransactionBuilderFactory,
    tsol: Sol.TransactionBuilderFactory,
    dot: Dot.TransactionBuilderFactory,
    tdot: Dot.TransactionBuilderFactory,
    near: Near.TransactionBuilderFactory,
    polygon: Polygon.TransactionBuilder,
    tpolygon: Polygon.TransactionBuilder,
    sui: Sui.TransactionBuilderFactory,
    tsui: Sui.TransactionBuilderFactory,
    tia: Tia.TransactionBuilderFactory,
    ttia: Tia.TransactionBuilderFactory,
    atom: Atom.TransactionBuilderFactory,
    tatom: Atom.TransactionBuilderFactory,
    osmo: Osmo.TransactionBuilderFactory,
    tosmo: Osmo.TransactionBuilderFactory,
    bld: Bld.TransactionBuilderFactory,
    tbld: Bld.TransactionBuilderFactory,
    hash: Hash.TransactionBuilderFactory,
    thash: Hash.TransactionBuilderFactory,
    sei: Sei.TransactionBuilderFactory,
    tsei: Sei.TransactionBuilderFactory,
    injective: Injective.TransactionBuilderFactory,
    tinjective: Injective.TransactionBuilderFactory,
    zeta: Zeta.TransactionBuilderFactory,
    tzeta: Zeta.TransactionBuilderFactory,
    islm: Islm.TransactionBuilderFactory,
    tislm: Islm.TransactionBuilderFactory,
    coreum: Coreum.TransactionBuilderFactory,
    tcoreum: Coreum.TransactionBuilderFactory,
    arbeth: Arbeth.TransactionBuilder,
    tarbeth: Arbeth.TransactionBuilder,
    opeth: Opeth.TransactionBuilder,
    topeth: Opeth.TransactionBuilder,
    ton: Ton.TransactionBuilder,
    tton: Ton.TransactionBuilder,
    zketh: zkEth.TransactionBuilder,
    tzketh: zkEth.TransactionBuilder,
    bera: Bera.TransactionBuilder,
    tbera: Bera.TransactionBuilder,
    rune: Rune.TransactionBuilderFactory,
    trune: Rune.TransactionBuilderFactory,
    oas: Oas.TransactionBuilder,
    toas: Oas.TransactionBuilder,
    coredao: Coredao.TransactionBuilder,
    tcoredao: Coredao.TransactionBuilder,
    apechain: Apechain.TransactionBuilder,
    tapechain: Apechain.TransactionBuilder,
    apt: Apt.TransactionBuilder,
    tapt: Apt.TransactionBuilder,
    tao: Tao.TransactionBuilderFactory,
    ttao: Tao.TransactionBuilderFactory,
    tempo: Tempo.Tip20TransactionBuilder,
    ttempo: Tempo.Tip20TransactionBuilder,
    icp: Icp.TransactionBuilder,
    ticp: Icp.TransactionBuilder,
    baby: Baby.TransactionBuilder,
    tbaby: Baby.TransactionBuilder,
    cronos: Cronos.TransactionBuilder,
    tcronos: Cronos.TransactionBuilder,
    initia: Initia.TransactionBuilder,
    tinitia: Initia.TransactionBuilder,
    asi: Asi.TransactionBuilder,
    tasi: Asi.TransactionBuilder,
    flr: Flr.TransactionBuilder,
    tflr: Flr.TransactionBuilder,
    sgb: Sgb.TransactionBuilder,
    tsgb: Sgb.TransactionBuilder,
    xdc: Xdc.TransactionBuilder,
    txdc: Xdc.TransactionBuilder,
    wemix: Wemix.TransactionBuilder,
    twemix: Wemix.TransactionBuilder,
    mon: Mon.TransactionBuilder,
    tmon: Mon.TransactionBuilder,
    world: World.TransactionBuilder,
    tworld: World.TransactionBuilder,
    soneium: Soneium.TransactionBuilder,
    tsoneium: Soneium.TransactionBuilder,
    tstt: Stt.TransactionBuilder,
    polyx: Polyx.TransactionBuilderFactory,
    tpolyx: Polyx.TransactionBuilderFactory,
    vet: Vet.TransactionBuilderFactory,
    tvet: Vet.TransactionBuilderFactory,
    canton: Canton.TransactionBuilderFactory,
    tcanton: Canton.TransactionBuilderFactory,
    flrp: sdk_coin_flrp_1.FlrPLib.TransactionBuilderFactory,
    tflrp: sdk_coin_flrp_1.FlrPLib.TransactionBuilderFactory,
};
const coinMessageBuilderFactoryMap = {
    ada: Ada.MessageBuilderFactory,
    tada: Ada.MessageBuilderFactory,
    sol: Sol.MessageBuilderFactory,
    tsol: Sol.MessageBuilderFactory,
};
statics_1.coins
    .filter((coin) => coin.features.includes(statics_1.CoinFeature.SHARED_EVM_MESSAGE_SIGNING) && !coin.isToken)
    .forEach((coin) => {
    coinMessageBuilderFactoryMap[coin.name] = Eth.MessageBuilderFactory;
});
statics_1.coins
    .filter((coin) => coin.features.includes(statics_1.CoinFeature.SHARED_EVM_SDK))
    .forEach((coin) => {
    coinBuilderMap[coin.name] = EvmCoin.TransactionBuilder;
});
statics_1.coins
    .filter((coin) => coin.features.includes(statics_1.CoinFeature.SHARED_COSMOS_SDK))
    .forEach((coin) => {
    coinBuilderMap[coin.name] = CosmosSharedCoin.TransactionBuilder;
});
/**
 * Get the list of coin tickers supported by this library.
 */
exports.supportedCoins = Object.keys(coinBuilderMap);
/**
 * Get a transaction builder for the given coin.
 *
 * @param {string} coinName One of the {@code supportedCoins}
 * @returns {any} An instance of a {@code TransactionBuilder}
 */
function getBuilder(coinName) {
    const builderClass = coinBuilderMap[coinName];
    if (!builderClass) {
        throw new sdk_core_1.BuildTransactionError(`Coin ${coinName} not supported`);
    }
    return new builderClass(statics_1.coins.get(coinName));
}
function getMessageBuilderFactory(coinName) {
    const messageBuilderFactoryClass = coinMessageBuilderFactoryMap[coinName];
    if (!messageBuilderFactoryClass) {
        throw new sdk_core_1.BuildMessageError(`Message builder factory for coin ${coinName} not supported`);
    }
    return new messageBuilderFactoryClass(statics_1.coins.get(coinName));
}
/**
 * Register a new coin instance with its builder factory
 *
 * @param {string} coinName coin name as it was registered in @bitgo/statics
 * @param {any} builderFactory the builder factory class for that coin
 * @returns {any} the factory instance for the registered coin
 */
function register(coinName, builderFactory) {
    const coinConfig = statics_1.coins.get(coinName);
    const factory = new builderFactory(coinConfig);
    // coinBuilderMap[coinName] = factory;
    coinBuilderMap[coinName] = builderFactory; // For now register the constructor function until reimplement getBuilder method
    return factory;
}
/**
 * Register a new coin instance with its message builder factory constructor.
 *
 * @param {string} coinName coin name as it was registered in @bitgo/statics
 * @param {any} messageBuilderFactory the message builder factory class for that coin
 * @returns {any} the message builder factory instance for the registered coin
 */
function registerMessageBuilderFactory(coinName, messageBuilderFactory) {
    const coinConfig = statics_1.coins.get(coinName);
    const factory = new messageBuilderFactory(coinConfig);
    coinMessageBuilderFactoryMap[coinName] = messageBuilderFactory;
    return factory;
}
/**
 * Verify a message against the given encoded payload.
 *
 * @param {string} coinName - The name of the coin.
 * @param {string} messageRaw - The raw message to verify.
 * @param {string} messageEncoded - The encoded message to verify against.
 * @param {MessageStandardType} messageStandardType - The type of message standard.
 * @param {Record<string, unknown>} [metadata] - Optional metadata for verification.
 * @returns {Promise<boolean>} - Returns true if the verification is successful, false otherwise.
 */
async function verifyMessage(coinName, messageRaw, messageEncoded, messageStandardType, metadata) {
    try {
        const messageBuilderFactory = getMessageBuilderFactory(coinName);
        const messageBuilder = messageBuilderFactory.getMessageBuilder(messageStandardType);
        messageBuilder.setPayload(messageRaw);
        const message = await messageBuilder.build();
        return await message.verifyEncodedPayload(messageEncoded, metadata);
    }
    catch (e) {
        console.error(`Error verifying message for coin ${coinName}:`, e);
        return false;
    }
}
