// @bitgo/sdk-coin-polygon/dist/src/polygonToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolygonToken = void 0;
/**
 * @prettier
 */
const abstract_eth_1 = require("@bitgo/abstract-eth");
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const lib_1 = require("./lib");
class PolygonToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, PolygonToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, PolygonToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(PolygonToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /**
     * Make a query to Polygonscan for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response from Polygonscan
     */
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].polygonscanApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].polygonscanBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    getFullName() {
        return 'Polygon Token';
    }
}
exports.PolygonToken = PolygonToken;
PolygonToken.coinNames = {
    Mainnet: 'polygon',
    Testnet: 'tpolygon',
};
