// @bitgo/blockapis/dist/src/BaseHttpClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseHttpClient = exports.ApiRequestError = void 0;
exports.mapSeries = mapSeries;
const superagent = require("superagent");
const logger_1 = require("@bitgo/logger");
class ApiRequestError extends Error {
    constructor(url, reason) {
        super(`error in API request: ${url}: ${reason}`);
        this.url = url;
        this.reason = reason;
    }
    static forRequest(req, reason) {
        return new ApiRequestError(req.toString(), reason);
    }
    static forResponse(res, reason) {
        return new ApiRequestError(res.header, reason);
    }
}
exports.ApiRequestError = ApiRequestError;
async function mapSeries(arr, f) {
    const results = [];
    for (const [index, value] of arr.entries()) {
        results.push(await f(value, index));
    }
    return results;
}
class BaseHttpClient {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
        if (baseUrl) {
            if (!baseUrl.startsWith('https://')) {
                throw new Error(`baseUrl must start with https://`);
            }
            if (baseUrl.endsWith(`/`)) {
                throw new Error(`baseUrl must not end with slash (/)`);
            }
        }
    }
    withBaseUrl(baseUrl) {
        return new BaseHttpClient(baseUrl);
    }
    async request(method, path, requestBody) {
        if (!path.startsWith(path)) {
            throw new Error(`path must start with '/' character`);
        }
        if (!this.baseUrl) {
            throw new Error(`no baseUrl set`);
        }
        const url = this.baseUrl + path;
        let response;
        try {
            response = await superagent(method, url).send(requestBody);
        }
        catch (e) {
            logger_1.logger.error(e);
            throw new ApiRequestError(url, e);
        }
        if (!response.ok) {
            throw new ApiRequestError(url, `status code ${response.statusCode}`);
        }
        let body;
        switch (response.type) {
            case 'application/json':
                body = response.body;
                break;
            case 'text/plain':
                body = response.text;
                break;
            default:
                throw new ApiRequestError(url, `unexpected response content-type ${response.type}`);
        }
        return {
            map(f) {
                try {
                    return f(body);
                }
                catch (e) {
                    throw new ApiRequestError(url, e);
                }
            },
        };
    }
    async get(path) {
        return this.request('get', path, undefined);
    }
}
exports.BaseHttpClient = BaseHttpClient;
