// @bitgo/blockapis/dist/src/CachingHttpClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachingHttpClient = void 0;
const url_1 = require("url");
const path = require("path");
const crypto = require("crypto");
const fs_1 = require("fs");
const BaseHttpClient_1 = require("./BaseHttpClient");
function stripApiTokens(uri) {
    const url = new url_1.URL(uri, 'http://localhost');
    url.searchParams.delete('key');
    return url.pathname + url.searchParams;
}
class CachingHttpClient {
    constructor(cacheDir, { client = new BaseHttpClient_1.BaseHttpClient(), isHttpEnabled = true } = {}) {
        this.cacheDir = cacheDir;
        this.client = client;
        this.isHttpEnabled = isHttpEnabled;
    }
    cachePath(p) {
        p = stripApiTokens(p).replace(/[^a-z0-9]/gi, '_');
        const hash = crypto.createHash('sha256').update(p).digest().toString('hex').slice(0, 8);
        return path.join(this.cacheDir, `${p}.${hash}.json`);
    }
    async readCache(path) {
        let data;
        try {
            data = await fs_1.promises.readFile(this.cachePath(path), 'utf8');
        }
        catch (e) {
            if (e.code === 'ENOENT') {
                return;
            }
            throw e;
        }
        return {
            map(f) {
                return f(JSON.parse(data));
            },
        };
    }
    async writeCache(path, data) {
        await fs_1.promises.writeFile(this.cachePath(path), JSON.stringify(data, null, 2), 'utf8');
    }
    async get(path) {
        const cached = await this.readCache(path);
        if (cached) {
            return cached;
        }
        if (!this.isHttpEnabled) {
            throw new Error(`networking disabled`);
        }
        const resp = await this.client.get(path);
        await this.writeCache(path, resp.map((v) => v));
        return resp;
    }
    withBaseUrl(baseUrl) {
        return new CachingHttpClient(this.cacheDir, {
            client: this.client.withBaseUrl(baseUrl),
            isHttpEnabled: this.isHttpEnabled,
        });
    }
}
exports.CachingHttpClient = CachingHttpClient;
