// @bitgo/sdk-coin-sol/dist/cjs/src/solToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolToken = void 0;
const sol_1 = require("./sol");
const statics_1 = require("@bitgo/statics");
class SolToken extends sol_1.Sol {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('sol') : statics_1.coins.get('tsol');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new SolToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.sol.tokens, ...statics_1.tokens.testnet.sol.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            const tokenConstructor = SolToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get tokenAddress() {
        return this.tokenConfig.tokenAddress;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    get contractAddress() {
        return this.tokenConfig.tokenAddress;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Solana Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return false;
    }
    /**
     * Flag for sending data along with transactions
     * @returns {boolean} True if okay to send tx data (CELO), false otherwise
     */
    transactionDataAllowed() {
        return false;
    }
}
exports.SolToken = SolToken;
