// @bitgo/sdk-coin-near/dist/src/lib/fungibleTokenTransferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FungibleTokenTransferBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const NearAPI = __importStar(require("near-api-js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
const contractCallWrapper_1 = require("./contractCallWrapper");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = __importDefault(require("./utils"));
class FungibleTokenTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.contractCallWrapper = new contractCallWrapper_1.ContractCallWrapper();
        this.contractCallWrapper.methodName = constants_1.FT_TRANSFER;
    }
    /**
     * Check if a transaction is a fungible token transfer
     *
     * @param {NearAPI.transactions.Action[]} actions near transaction actions
     * @returns {Boolean} true if more than 1 action present or first action method name is ft transfer
     */
    static isFungibleTokenTransferTransaction(actions) {
        return actions.length > 1 || actions[0].functionCall?.methodName === constants_1.FT_TRANSFER;
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const action of tx.nearTransaction.actions) {
            if (action.functionCall && action.functionCall.methodName === constants_1.FT_TRANSFER) {
                this.contractCallWrapper.deposit = action.functionCall.deposit.toString();
                this.contractCallWrapper.gas = action.functionCall.gas.toString();
            }
        }
    }
    /**
     * Sets the gas of this transaction.
     *
     * @param {String} gas the gas of this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    gas(gas) {
        this.validateValue(new bignumber_js_1.default(gas));
        this.contractCallWrapper.gas = gas;
        return this;
    }
    /**
     * Sets the deposit of at-least 1 yoctoNear
     *
     * @param {string} deposit the deposit in the minimum unit (1 Near = 1e24 yoctoNear) of this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    deposit(deposit) {
        this.validateValue(new bignumber_js_1.default(deposit));
        this.contractCallWrapper.deposit = deposit;
        return this;
    }
    /**
     * Sets the actual receiver account id inside args
     *
     * @param accountId the receiver account id
     */
    ftReceiverId(accountId) {
        utils_1.default.isValidAddress(accountId);
        this.contractCallWrapper.args = { receiver_id: accountId };
        return this;
    }
    /**
     * Sets the ft amount to be transferred
     *
     * @param amount the amount of fungible token to be transferred
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this.contractCallWrapper.args = { amount };
        return this;
    }
    /**
     * Sets the optional memo for the transfer
     *
     * @param memo
     */
    memo(memo) {
        this.contractCallWrapper.args = { memo };
        return this;
    }
    /**
     * Sets the storage deposit action
     *
     * @param {StorageDepositInput} input contains the deposit value, gas and optional account id
     * if account id is not provided then it is self transfer
     */
    addStorageDeposit(input) {
        const methodName = constants_1.STORAGE_DEPOSIT;
        (0, assert_1.default)(input.deposit, new sdk_core_1.BuildTransactionError('deposit is required before building storage deposit transfer'));
        (0, assert_1.default)(input.gas, new sdk_core_1.BuildTransactionError('gas is required before building fungible token transfer'));
        const args = input.accountId ? { account_id: input.accountId } : {};
        const action = NearAPI.transactions.functionCall(methodName, args, input.gas, input.deposit);
        super.action(action);
    }
    /** @inheritdoc */
    async buildImplementation() {
        const { methodName, args, gas, deposit } = this.contractCallWrapper.getParams();
        (0, assert_1.default)(gas, new sdk_core_1.BuildTransactionError('gas is required before building fungible token transfer'));
        (0, assert_1.default)(deposit, new sdk_core_1.BuildTransactionError('deposit is required before building fungible token transfer'));
        if (!this._actions ||
            this._actions.length === 0 ||
            (this._actions.length === 1 && this._actions[0].functionCall?.methodName !== methodName)) {
            super.action(NearAPI.transactions.functionCall(methodName, args, BigInt(gas), BigInt(deposit)));
        }
        const tx = await super.buildImplementation();
        tx.setTransactionType(sdk_core_1.TransactionType.Send);
        return tx;
    }
}
exports.FungibleTokenTransferBuilder = FungibleTokenTransferBuilder;
