// @bitgo/sdk-coin-near/dist/src/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.TransactionBuilderFactory = exports.StorageDepositTransferBuilder = exports.FungibleTokenTransferBuilder = exports.MetaPoolWithdrawBuilder = exports.StakingWithdrawBuilder = exports.StakingDeactivateBuilder = exports.StakingActivateBuilder = exports.TransferBuilder = exports.TransactionBuilder = exports.Transaction = exports.KeyPair = void 0;
const Utils = __importStar(require("./utils"));
exports.Utils = Utils;
var keyPair_1 = require("./keyPair");
Object.defineProperty(exports, "KeyPair", { enumerable: true, get: function () { return keyPair_1.KeyPair; } });
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return transaction_1.Transaction; } });
var transactionBuilder_1 = require("./transactionBuilder");
Object.defineProperty(exports, "TransactionBuilder", { enumerable: true, get: function () { return transactionBuilder_1.TransactionBuilder; } });
var transferBuilder_1 = require("./transferBuilder");
Object.defineProperty(exports, "TransferBuilder", { enumerable: true, get: function () { return transferBuilder_1.TransferBuilder; } });
var stakingActivateBuilder_1 = require("./stakingActivateBuilder");
Object.defineProperty(exports, "StakingActivateBuilder", { enumerable: true, get: function () { return stakingActivateBuilder_1.StakingActivateBuilder; } });
var stakingDeactivateBuilder_1 = require("./stakingDeactivateBuilder");
Object.defineProperty(exports, "StakingDeactivateBuilder", { enumerable: true, get: function () { return stakingDeactivateBuilder_1.StakingDeactivateBuilder; } });
var stakingWithdrawBuilder_1 = require("./stakingWithdrawBuilder");
Object.defineProperty(exports, "StakingWithdrawBuilder", { enumerable: true, get: function () { return stakingWithdrawBuilder_1.StakingWithdrawBuilder; } });
var metaPoolWithdrawBuilder_1 = require("./metaPoolWithdrawBuilder");
Object.defineProperty(exports, "MetaPoolWithdrawBuilder", { enumerable: true, get: function () { return metaPoolWithdrawBuilder_1.MetaPoolWithdrawBuilder; } });
var fungibleTokenTransferBuilder_1 = require("./fungibleTokenTransferBuilder");
Object.defineProperty(exports, "FungibleTokenTransferBuilder", { enumerable: true, get: function () { return fungibleTokenTransferBuilder_1.FungibleTokenTransferBuilder; } });
var storageDepositTransferBuilder_1 = require("./storageDepositTransferBuilder");
Object.defineProperty(exports, "StorageDepositTransferBuilder", { enumerable: true, get: function () { return storageDepositTransferBuilder_1.StorageDepositTransferBuilder; } });
var transactionBuilderFactory_1 = require("./transactionBuilderFactory");
Object.defineProperty(exports, "TransactionBuilderFactory", { enumerable: true, get: function () { return transactionBuilderFactory_1.TransactionBuilderFactory; } });
