// @bitgo/sdk-coin-near/dist/src/lib/storageDepositTransferBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageDepositTransferBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const NearAPI = __importStar(require("near-api-js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const contractCallWrapper_1 = require("./contractCallWrapper");
const transactionBuilder_1 = require("./transactionBuilder");
const constants_1 = require("./constants");
const utils_1 = __importDefault(require("./utils"));
class StorageDepositTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.contractCallWrapper = new contractCallWrapper_1.ContractCallWrapper();
        this.contractCallWrapper.methodName = constants_1.STORAGE_DEPOSIT;
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const action of tx.nearTransaction.actions) {
            if (action.functionCall && action.functionCall.methodName === constants_1.STORAGE_DEPOSIT) {
                this.contractCallWrapper.deposit = action.functionCall.deposit.toString();
                this.contractCallWrapper.gas = action.functionCall.gas.toString();
                break;
            }
        }
    }
    /**
     * Sets the gas of this transaction.
     *
     * @param {String} gas the gas of this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    gas(gas) {
        this.validateValue(new bignumber_js_1.default(gas));
        this.contractCallWrapper.gas = gas;
        return this;
    }
    /**
     * Sets the deposit of at-least 1 yoctoNear
     *
     * @param {string} deposit the deposit in the minimum unit (1 Near = 1e24 yoctoNear) of this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    deposit(deposit) {
        this.validateValue(new bignumber_js_1.default(deposit));
        this.contractCallWrapper.deposit = deposit;
        return this;
    }
    /**
     * Sets the actual receiver account id inside args
     *
     * @param accountId the receiver account id
     */
    beneficiaryId(accountId) {
        utils_1.default.isValidAddress(accountId);
        this.contractCallWrapper.args = { account_id: accountId };
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        const { methodName, args, gas, deposit } = this.contractCallWrapper.getParams();
        (0, assert_1.default)(gas, new sdk_core_1.BuildTransactionError('gas is required before building fungible token transfer'));
        (0, assert_1.default)(deposit, new sdk_core_1.BuildTransactionError('deposit is required before building fungible token transfer'));
        if (!this._actions || !this._actions.length) {
            super.action(NearAPI.transactions.functionCall(methodName, args, BigInt(gas), BigInt(deposit)));
        }
        const tx = await super.buildImplementation();
        tx.setTransactionType(sdk_core_1.TransactionType.StorageDeposit);
        return tx;
    }
}
exports.StorageDepositTransferBuilder = StorageDepositTransferBuilder;
