// @bitgo/sdk-coin-near/dist/src/lib/transactionBuilderFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transferBuilder_1 = require("./transferBuilder");
const walletInitializationBuilder_1 = require("./walletInitializationBuilder");
const transaction_1 = require("./transaction");
const stakingActivateBuilder_1 = require("./stakingActivateBuilder");
const stakingDeactivateBuilder_1 = require("./stakingDeactivateBuilder");
const stakingWithdrawBuilder_1 = require("./stakingWithdrawBuilder");
const metaPoolWithdrawBuilder_1 = require("./metaPoolWithdrawBuilder");
const fungibleTokenTransferBuilder_1 = require("./fungibleTokenTransferBuilder");
const storageDepositTransferBuilder_1 = require("./storageDepositTransferBuilder");
const constants_1 = require("./constants");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    from(raw) {
        try {
            const tx = new transaction_1.Transaction(this._coinConfig);
            tx.fromRawTransaction(raw);
            switch (tx.type) {
                case sdk_core_1.TransactionType.Send:
                    if (fungibleTokenTransferBuilder_1.FungibleTokenTransferBuilder.isFungibleTokenTransferTransaction(tx.nearTransaction.actions)) {
                        return this.getFungibleTokenTransferBuilder(tx);
                    }
                    return this.getTransferBuilder(tx);
                case sdk_core_1.TransactionType.WalletInitialization:
                    return this.getWalletInitializationBuilder(tx);
                case sdk_core_1.TransactionType.StakingActivate:
                    return this.getStakingActivateBuilder(tx);
                case sdk_core_1.TransactionType.StakingDeactivate:
                    return this.getStakingDeactivateBuilder(tx);
                case sdk_core_1.TransactionType.StakingWithdraw: {
                    const methodName = tx.nearTransaction.actions[0]?.functionCall?.methodName;
                    if (methodName === constants_1.StakingContractMethodNames.WithdrawAll) {
                        return this.getMetaPoolWithdrawBuilder(tx);
                    }
                    return this.getStakingWithdrawBuilder(tx);
                }
                case sdk_core_1.TransactionType.StorageDeposit:
                    return this.getStorageDepositTransferBuilder(tx);
                default:
                    throw new sdk_core_1.InvalidTransactionError('unsupported transaction');
            }
        }
        catch (e) {
            throw e;
        }
    }
    /** @inheritdoc */
    getWalletInitializationBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new walletInitializationBuilder_1.WalletInitializationBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getTransferBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    getStakingActivateBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new stakingActivateBuilder_1.StakingActivateBuilder(this._coinConfig));
    }
    getStakingDeactivateBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new stakingDeactivateBuilder_1.StakingDeactivateBuilder(this._coinConfig));
    }
    getStakingWithdrawBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new stakingWithdrawBuilder_1.StakingWithdrawBuilder(this._coinConfig));
    }
    getMetaPoolWithdrawBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new metaPoolWithdrawBuilder_1.MetaPoolWithdrawBuilder(this._coinConfig));
    }
    getFungibleTokenTransferBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new fungibleTokenTransferBuilder_1.FungibleTokenTransferBuilder(this._coinConfig));
    }
    getStorageDepositTransferBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new storageDepositTransferBuilder_1.StorageDepositTransferBuilder(this._coinConfig));
    }
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    static initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
