// @bitgo/sdk-coin-near/dist/src/lib/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const bs58_1 = __importDefault(require("bs58"));
const js_sha256_1 = require("js-sha256");
const nearAPI = __importStar(require("near-api-js"));
const transaction_1 = require("near-api-js/lib/transaction");
const crypto_1 = require("@near-js/crypto");
const transactions_1 = require("@near-js/transactions");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
const keyPair_1 = require("./keyPair");
const utils_1 = __importDefault(require("./utils"));
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
    }
    get nearTransaction() {
        return this._nearTransaction;
    }
    set nearTransaction(tx) {
        this._nearTransaction = tx;
        this._id = utils_1.default.base58Encode(this.getTransactionHash());
    }
    /** @inheritdoc */
    canSign(key) {
        try {
            new keyPair_1.KeyPair({ prv: key.key });
            return true;
        }
        catch {
            return false;
        }
    }
    /** @inheritdoc */
    toBroadcastFormat() {
        if (!this._nearTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        const txSeralized = this._nearSignedTransaction
            ? Buffer.from(this._nearSignedTransaction.encode()).toString('base64')
            : Buffer.from(this._nearTransaction.encode()).toString('base64');
        return txSeralized;
    }
    /** @inheritdoc */
    toJson() {
        if (!this._nearTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        const parsedActions = [];
        const actions = this._nearTransaction.actions;
        actions.map((action) => {
            let parsedAction = {};
            if (action.enum === 'transfer' && action.transfer) {
                parsedAction = { transfer: action.transfer };
            }
            else if (action.enum === 'functionCall' && action.functionCall) {
                const functionCall = action.functionCall;
                parsedAction = {
                    functionCall: {
                        methodName: functionCall.methodName,
                        args: JSON.parse(Buffer.from(functionCall.args).toString()),
                        gas: functionCall.gas.toString(),
                        deposit: functionCall.deposit.toString(),
                    },
                };
            }
            parsedActions.push(parsedAction);
        });
        let signature;
        if (this._nearSignedTransaction?.signature?.ed25519Signature) {
            signature = {
                keyType: crypto_1.KeyType.ED25519,
                data: new Uint8Array(this._nearSignedTransaction.signature.ed25519Signature.data),
            };
        }
        let publicKey;
        if (this._nearTransaction.publicKey?.ed25519Key) {
            const rawBytes = new Uint8Array(this._nearTransaction.publicKey.ed25519Key.data);
            const encoded = nearAPI.utils.serialize.base_encode(rawBytes);
            publicKey = `ed25519:${encoded}`;
        }
        return {
            id: this._id,
            signerId: this._nearTransaction.signerId,
            publicKey: publicKey,
            nonce: this._nearTransaction.nonce,
            receiverId: this._nearTransaction.receiverId,
            actions: parsedActions,
            signature: signature ? signature : undefined,
        };
    }
    /**
     * Set the transaction type.
     *
     * @param {TransactionType} transactionType The transaction type to be set.
     */
    setTransactionType(transactionType) {
        this._type = transactionType;
    }
    /**
     * Sets this transaction payload
     *
     * @param rawTx
     */
    fromRawTransaction(rawTx) {
        const bufferRawTransaction = constants_1.HEX_REGEX.test(rawTx) ? Buffer.from(rawTx, 'hex') : Buffer.from(rawTx, 'base64');
        try {
            const signedTx = nearAPI.utils.serialize.deserialize(nearAPI.transactions.SCHEMA.SignedTransaction, bufferRawTransaction);
            signedTx.transaction.actions = signedTx.transaction.actions.map((a) => {
                const action = new transactions_1.Action(a);
                switch (action.enum) {
                    case 'transfer': {
                        if (action.transfer?.deposit) {
                            return (0, transaction_1.transfer)(BigInt(action.transfer.deposit));
                        }
                        break;
                    }
                    case 'functionCall': {
                        if (action.functionCall) {
                            return (0, transaction_1.functionCall)(action.functionCall.methodName, new Uint8Array(action.functionCall.args), BigInt(action.functionCall.gas), BigInt(action.functionCall.deposit));
                        }
                        break;
                    }
                }
                return action;
            });
            this._nearSignedTransaction = signedTx;
            this._nearTransaction = signedTx.transaction;
            this._id = utils_1.default.base58Encode(this.getTransactionHash());
        }
        catch (e) {
            try {
                const unsignedTx = nearAPI.utils.serialize.deserialize(nearAPI.transactions.SCHEMA.Transaction, bufferRawTransaction);
                unsignedTx.actions = unsignedTx.actions.map((a) => {
                    const action = new transactions_1.Action(a);
                    switch (action.enum) {
                        case 'transfer': {
                            if (action.transfer?.deposit) {
                                return (0, transaction_1.transfer)(BigInt(action.transfer.deposit));
                            }
                            break;
                        }
                        case 'functionCall': {
                            if (action.functionCall) {
                                return (0, transaction_1.functionCall)(action.functionCall.methodName, new Uint8Array(action.functionCall.args), BigInt(action.functionCall.gas), BigInt(action.functionCall.deposit));
                            }
                            break;
                        }
                    }
                    return action;
                });
                this._nearTransaction = unsignedTx;
                this._id = utils_1.default.base58Encode(this.getTransactionHash());
            }
            catch (e) {
                throw new sdk_core_1.InvalidTransactionError('unable to build transaction from raw');
            }
        }
        this.loadInputsAndOutputs();
    }
    /**
     * Sign this transaction
     *
     * @param {KeyPair} signer key
     */
    sign(signer) {
        if (!this._nearTransaction) {
            throw new sdk_core_1.InvalidTransactionError('empty transaction to sign');
        }
        const serializedTxHash = this.getTransactionHash();
        const signature = signer.signMessageinUint8Array(serializedTxHash);
        this._nearSignedTransaction = new nearAPI.transactions.SignedTransaction({
            transaction: this._nearTransaction,
            signature: new nearAPI.transactions.Signature({
                keyType: this._nearTransaction.publicKey.keyType,
                data: signature,
            }),
        });
        this.loadInputsAndOutputs();
    }
    /**
     * set transaction type by staking contract method names.
     * @param methodName method name to match and set the transaction type
     */
    setTypeByFunctionCall(methodName) {
        switch (methodName) {
            case constants_1.StakingContractMethodNames.DepositAndStake:
                this.setTransactionType(sdk_core_1.TransactionType.StakingActivate);
                break;
            case constants_1.StakingContractMethodNames.Unstake:
                this.setTransactionType(sdk_core_1.TransactionType.StakingDeactivate);
                break;
            case constants_1.StakingContractMethodNames.Withdraw:
            case constants_1.StakingContractMethodNames.WithdrawAll:
                this.setTransactionType(sdk_core_1.TransactionType.StakingWithdraw);
                break;
            case constants_1.FT_TRANSFER:
                this.setTransactionType(sdk_core_1.TransactionType.Send);
                break;
            case constants_1.STORAGE_DEPOSIT:
                this.setTransactionType(sdk_core_1.TransactionType.StorageDeposit);
                break;
        }
    }
    /**
     * Check if method is allowed on Near account-lib implementation.
     * This method should check on all contracts added to Near.
     * @param methodName contract call method name to check if it's allowed.
     */
    validateMethodAllowed(methodName) {
        const allowedMethods = [...Object.values(constants_1.StakingContractMethodNames), ...constants_1.FUNGIBLE_TOKEN_RELATED_METHODS];
        if (!allowedMethods.includes(methodName)) {
            throw new sdk_core_1.InvalidTransactionError('unsupported function call in raw transaction');
        }
    }
    /**
     * Check if valid methods are present for batching of actions
     *
     * @param {TxAction[]} actions list of near transaction actions
     * @returns {void}
     */
    validateBatchingMethods(actions) {
        actions.every((action) => {
            if (action.enum !== 'functionCall' || !action.functionCall) {
                throw new sdk_core_1.InvalidTransactionError('invalid action for batching');
            }
        });
        if (actions[0].functionCall?.methodName !== constants_1.STORAGE_DEPOSIT ||
            actions[1].functionCall?.methodName !== constants_1.FT_TRANSFER) {
            throw new sdk_core_1.InvalidTransactionError('invalid action sequence: expected storage_deposit followed by ft_transfer');
        }
    }
    /**
     * Build input and output field for this transaction
     *
     */
    loadInputsAndOutputs() {
        if (!this._nearTransaction.actions || !this._nearTransaction.actions.length) {
            throw new sdk_core_1.InvalidTransactionError('no actions in raw transaction');
        }
        // TODO: modify this for send-many support
        // currently only storage deposit + ft transfer are allowed for batching
        if (this._nearTransaction.actions.length > 2) {
            throw new sdk_core_1.InvalidTransactionError('too many actions in raw transaction');
        }
        // check for correct sequence of actions
        if (this._nearTransaction.actions.length === 2) {
            this.validateBatchingMethods(this._nearTransaction.actions);
        }
        const action = this._nearTransaction.actions[0];
        const actions = this._nearTransaction.actions;
        switch (action.enum) {
            case 'transfer':
                this.setTransactionType(sdk_core_1.TransactionType.Send);
                break;
            case 'functionCall':
                if (action.functionCall) {
                    const methodName = action.functionCall.methodName;
                    this.validateMethodAllowed(methodName);
                    this.setTypeByFunctionCall(methodName);
                }
                break;
            default:
                throw new sdk_core_1.InvalidTransactionError('unsupported action in raw transaction');
        }
        // if there are 2 actions, we know for sure that it is storage deposit + ft transfer
        if (actions.length === 2) {
            this.setTransactionType(sdk_core_1.TransactionType.Send);
        }
        const outputs = [];
        const inputs = [];
        actions.map((action) => {
            switch (this.type) {
                case sdk_core_1.TransactionType.Send:
                    if (action.transfer) {
                        const amount = action.transfer.deposit.toString();
                        inputs.push({
                            address: this._nearTransaction.signerId,
                            value: amount,
                            coin: this._coinConfig.name,
                        });
                        outputs.push({
                            address: this._nearTransaction.receiverId,
                            value: amount,
                            coin: this._coinConfig.name,
                        });
                    }
                    else if (action.functionCall) {
                        if (action.functionCall.methodName === 'ft_transfer') {
                            const parsedArgs = JSON.parse(Buffer.from(action.functionCall.args).toString());
                            inputs.push({
                                address: this._nearTransaction.signerId,
                                value: parsedArgs.amount,
                                coin: this._coinConfig.name,
                            });
                            outputs.push({
                                address: parsedArgs.receiver_id,
                                value: parsedArgs.amount,
                                coin: this._coinConfig.name,
                            });
                        }
                    }
                    break;
                case sdk_core_1.TransactionType.StakingActivate:
                    if (action.functionCall) {
                        const stakingAmount = action.functionCall.deposit.toString();
                        inputs.push({
                            address: this._nearTransaction.signerId,
                            value: stakingAmount,
                            coin: this._coinConfig.name,
                        });
                        outputs.push({
                            address: this._nearTransaction.receiverId,
                            value: stakingAmount,
                            coin: this._coinConfig.name,
                        });
                    }
                    break;
                case sdk_core_1.TransactionType.StakingWithdraw:
                    if (action.functionCall) {
                        const methodName = action.functionCall.methodName;
                        if (methodName === constants_1.StakingContractMethodNames.WithdrawAll) {
                            // withdraw_all has no amount arg; amount is determined on-chain
                            break;
                        }
                        const stakingWithdrawAmount = JSON.parse(Buffer.from(action.functionCall.args).toString()).amount;
                        inputs.push({
                            address: this._nearTransaction.receiverId,
                            value: stakingWithdrawAmount,
                            coin: this._coinConfig.name,
                        });
                        outputs.push({
                            address: this._nearTransaction.signerId,
                            value: stakingWithdrawAmount,
                            coin: this._coinConfig.name,
                        });
                    }
                    break;
                case sdk_core_1.TransactionType.StorageDeposit:
                    break;
            }
        });
        this._outputs = outputs;
        this._inputs = inputs;
    }
    /**
     * Returns a complete explanation for a transfer transaction
     * @param {TxData} json The transaction data in json format
     * @param {TransactionExplanation} explanationResult The transaction explanation to be completed
     * @returns {TransactionExplanation}
     */
    explainTransferTransaction(json, explanationResult) {
        return {
            ...explanationResult,
            outputAmount: json.actions[0].transfer?.deposit.toString() || '',
            outputs: [
                {
                    address: json.receiverId,
                    amount: json.actions[0].transfer?.deposit.toString() || '',
                },
            ],
        };
    }
    /**
     * Returns a complete explanation for a staking activate transaction
     * @param {TxData} json The transaction data in json format
     * @param {TransactionExplanation} explanationResult The transaction explanation to be completed
     * @returns {TransactionExplanation}
     */
    explainStakingActivateTransaction(json, explanationResult) {
        return {
            ...explanationResult,
            outputAmount: json.actions[0].functionCall?.deposit.toString() || '',
            outputs: [
                {
                    address: json.receiverId,
                    amount: json.actions[0].functionCall?.deposit.toString() || '',
                },
            ],
        };
    }
    /**
     * Returns a complete explanation for a staking withdraw transaction
     * @param {TxData} json The transaction data in json format
     * @param {TransactionExplanation} explanationResult The transaction explanation to be completed
     * @returns {TransactionExplanation}
     */
    explainStakingWithdrawTransaction(json, explanationResult) {
        const methodName = json.actions[0].functionCall?.methodName;
        if (methodName === constants_1.StakingContractMethodNames.WithdrawAll) {
            // withdraw_all has no amount arg; amount is resolved on-chain
            return {
                ...explanationResult,
                outputAmount: '0',
                outputs: [
                    {
                        address: json.signerId,
                        amount: '0',
                    },
                ],
            };
        }
        const amount = json.actions[0].functionCall?.args.amount;
        return {
            ...explanationResult,
            outputAmount: amount,
            outputs: [
                {
                    address: json.signerId,
                    amount: amount,
                },
            ],
        };
    }
    /**
     * Calculates the total fungible token amount & total native near amount
     *
     * @param {Action[]} actions the list of delegate transaction actions
     * @returns {String} either native near amount or fungible token amount
     */
    calculateTotalOutputAmount(actions) {
        let hasFtTransfer = false;
        let hasStorageDeposit = false;
        let totalTokenAmount = BigInt(0);
        let totalNearDeposit = BigInt(0);
        for (const action of actions) {
            if (action.functionCall) {
                const functionCall = action.functionCall;
                const methodName = functionCall.methodName;
                const args = functionCall.args;
                const deposit = BigInt(functionCall.deposit);
                if (methodName === constants_1.FT_TRANSFER) {
                    hasFtTransfer = true;
                    const amountStr = args['amount'];
                    if (args.amount) {
                        totalTokenAmount += BigInt(amountStr);
                    }
                }
                if (methodName === constants_1.STORAGE_DEPOSIT) {
                    hasStorageDeposit = true;
                    totalNearDeposit += deposit;
                }
            }
        }
        if (hasFtTransfer) {
            return totalTokenAmount.toString();
        }
        else if (hasStorageDeposit) {
            return totalNearDeposit.toString();
        }
        return '';
    }
    /**
     * Returns a complete explanation for a token transfer transaction
     * @param {TxData} json The transaction data in json format
     * @param {TransactionExplanation} explanationResult The transaction explanation to be completed
     * @returns {TransactionExplanation}
     */
    explainTokenTransferTransaction(json, explanationResult) {
        const actions = json.actions;
        const outputAmount = this.calculateTotalOutputAmount(actions);
        const outputs = [];
        actions.map((action) => {
            if (action.functionCall) {
                const functionCall = action.functionCall;
                if (functionCall.methodName === constants_1.FT_TRANSFER) {
                    const amountStr = functionCall.args['amount'];
                    const receiverId = functionCall.args['receiver_id'];
                    // in ft transfer, the outer receiver id will be contract address of the token
                    const tokenName = utils_1.default.findTokenNameFromContractAddress(json.receiverId);
                    const output = {
                        address: receiverId,
                        amount: amountStr,
                    };
                    if (tokenName) {
                        output.tokenName = tokenName;
                    }
                    outputs.push(output);
                }
            }
        });
        return {
            ...explanationResult,
            outputAmount: outputAmount,
            outputs: outputs,
        };
    }
    /**
     * Returns a complete explanation for a storage deposit transaction
     * @param {TxData} json The transaction data in json format
     * @param {TransactionExplanation} explanationResult The transaction explanation to be completed
     * @returns {TransactionExplanation}
     */
    explainStorageDepositTransaction(json, explanationResult) {
        const actions = json.actions;
        const outputAmount = this.calculateTotalOutputAmount(actions);
        const outputs = [];
        actions.map((action) => {
            if (action.functionCall) {
                const functionCall = action.functionCall;
                if (functionCall.methodName === constants_1.STORAGE_DEPOSIT) {
                    const receiverId = functionCall.args && functionCall.args['account_id']
                        ? functionCall.args['account_id']
                        : json.signerId;
                    // in storage deposit, the outer receiver id will be contract address of the token
                    const tokenName = utils_1.default.findTokenNameFromContractAddress(json.receiverId);
                    const output = {
                        address: receiverId,
                        amount: functionCall.deposit,
                    };
                    if (tokenName) {
                        output.tokenName = tokenName;
                    }
                    outputs.push(output);
                }
            }
        });
        return {
            ...explanationResult,
            outputAmount: outputAmount,
            outputs: outputs,
        };
    }
    /** @inheritdoc */
    explainTransaction() {
        const result = this.toJson();
        const displayOrder = ['outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'type'];
        const outputs = [];
        const explanationResult = {
            // txhash used to identify the transactions
            id: result.id || '',
            displayOrder,
            outputAmount: '0',
            changeAmount: '0',
            changeOutputs: [],
            outputs,
            fee: { fee: '' },
            type: this.type,
        };
        switch (this.type) {
            case sdk_core_1.TransactionType.Send:
                if (result.actions.length > 1 || result.actions[0].functionCall) {
                    return this.explainTokenTransferTransaction(result, explanationResult);
                }
                return this.explainTransferTransaction(result, explanationResult);
            case sdk_core_1.TransactionType.StakingActivate:
                return this.explainStakingActivateTransaction(result, explanationResult);
            case sdk_core_1.TransactionType.StakingDeactivate:
                return explanationResult;
            case sdk_core_1.TransactionType.StakingWithdraw:
                return this.explainStakingWithdrawTransaction(result, explanationResult);
            case sdk_core_1.TransactionType.StorageDeposit:
                return this.explainStorageDepositTransaction(result, explanationResult);
            default:
                throw new sdk_core_1.InvalidTransactionError('Transaction type not supported');
        }
    }
    getTransactionHash() {
        const serializedTx = nearAPI.utils.serialize.serialize(nearAPI.transactions.SCHEMA.Transaction, this._nearTransaction);
        return new Uint8Array(js_sha256_1.sha256.array(serializedTx));
    }
    get signablePayload() {
        if (!this._nearTransaction) {
            throw new sdk_core_1.InvalidTransactionError('empty transaction');
        }
        return Buffer.from(this.getTransactionHash());
    }
    /**
     * Constructs a signed payload using construct.signTx
     * This method will be called during the build step if a TSS signature
     * is added and will set the signTransaction which is the txHex that will be broadcasted
     * As well as add the signature used to sign to the signature array in hex format
     *
     * @param {Buffer} signature The signature to be added to a near transaction
     */
    constructSignedPayload(signature) {
        this._nearSignedTransaction = new nearAPI.transactions.SignedTransaction({
            transaction: this._nearTransaction,
            signature: new nearAPI.transactions.Signature({
                keyType: this._nearTransaction.publicKey.keyType,
                data: signature,
            }),
        });
        this.loadInputsAndOutputs();
    }
    /** @inheritdoc **/
    get signature() {
        const signatures = [];
        if (this._nearSignedTransaction && this._nearSignedTransaction.signature.ed25519Signature) {
            signatures.push(bs58_1.default.encode(this._nearSignedTransaction.signature.ed25519Signature.data));
        }
        return signatures;
    }
}
exports.Transaction = Transaction;
