// @bitgo/sdk-coin-near/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const bs58_1 = __importDefault(require("bs58"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const keyPair_1 = require("./keyPair");
class Utils {
    /** @inheritdoc */
    isValidAddress(address) {
        return this.isValidAccountId(address);
    }
    /** @inheritdoc */
    isValidBlockId(hash) {
        return (0, sdk_core_1.isBase58)(hash, 32);
    }
    /** @inheritdoc */
    isValidPrivateKey(key) {
        return this.isValidKey(key);
    }
    /** @inheritdoc */
    isValidPublicKey(pubKey) {
        return this.isValidAccountId(pubKey);
    }
    /**
     * Check if the key have the correct format
     *
     * @param {string} key - string to be checked
     * @return {boolean} true if the string have the correct format otherwise return false
     */
    isValidKey(key) {
        try {
            new keyPair_1.KeyPair({ prv: key });
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Check if the account have the correct format
     *
     * @param {string} accountId - string to be checked
     * @return {boolean} if the string can have the correct format and match the expected length
     * or can create a new keyPair
     */
    isValidAccountId(accountId) {
        if (!accountId) {
            return false;
        }
        return ((/^(([a-z\d]+[\-_])*[a-z\d]+\.)*([a-z\d]+[\-_])*[a-z\d]+$/.test(accountId) &&
            accountId.length >= 2 &&
            accountId.length <= 64) ||
            (0, sdk_core_1.isBase58)(accountId, 32));
    }
    /** @inheritdoc */
    isValidSignature(signature) {
        return (0, sdk_core_1.isBase58)(signature, 64);
    }
    /** @inheritdoc */
    isValidTransactionId(txId) {
        return (0, sdk_core_1.isBase58)(txId, 32);
    }
    base58Encode(value) {
        return bs58_1.default.encode(value);
    }
    /**
     * Check if base58 decoded string is equal to length
     *
     * @param {string} value - string to be checked
     * @param {number} length - expected decoded length
     * @return {boolean} if the string can be decoded as base58 and match the expected length
     */
    isBase58(value, length) {
        try {
            return !!value && bs58_1.default.decode(value).length === length;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Find the bitgo token name using contract address
     *
     * @param {String} contractAddress the token contract address
     * @returns {String} token name
     */
    findTokenNameFromContractAddress(contractAddress) {
        const token = statics_1.coins
            .filter((coin) => coin instanceof statics_1.Nep141Token && coin.contractAddress === contractAddress)
            .map((coin) => coin);
        return token ? token[0].name : undefined;
    }
    /**
     * Find the token instance using the bitgo token name
     *
     * @param {String} tokenName the bitgo name of the token
     * @returns {Nep141Token|undefined} token instance if found
     */
    getTokenInstanceFromTokenName(tokenName) {
        const token = statics_1.coins
            .filter((coin) => coin instanceof statics_1.Nep141Token && coin.name === tokenName)
            .map((coin) => coin);
        return token ? token[0] : undefined;
    }
    /**
     * Convert from raw gas units to yocto Near value
     *
     * @param {String} gasUnits - raw gas units (e.g. "30000000000000" = 30TGas)
     * @returns {String} value in yoctoNear as a string
     */
    convertGasUnitsToYoctoNear(gasUnits) {
        const YOCTO_PER_NEAR = new bignumber_js_1.default('1e24');
        const NEAR_PER_TGAS = new bignumber_js_1.default('0.0001'); // 1 TGas = 0.0001 Near
        const GAS_UNITS_PER_TGAS = new bignumber_js_1.default('1e12'); // 1 TGas = 1e12 gas units
        const gas = new bignumber_js_1.default(gasUnits);
        const tgas = gas.dividedBy(GAS_UNITS_PER_TGAS);
        const nearCost = tgas.multipliedBy(NEAR_PER_TGAS);
        const yoctoCost = nearCost.multipliedBy(YOCTO_PER_NEAR);
        return yoctoCost.integerValue(bignumber_js_1.default.ROUND_FLOOR).toString();
    }
    /**
     * Convert the given record into base64 string
     *
     * @param {Record<String, String>} data the input in the form of a map
     * @returns {String} base64 encoded string
     */
    convertToBase64(data) {
        return Buffer.from(JSON.stringify(data)).toString('base64');
    }
}
exports.Utils = Utils;
const utils = new Utils();
exports.default = utils;
