// @bitgo/sdk-coin-coredao/dist/src/coredaoToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoredaoToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class CoredaoToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, CoredaoToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, CoredaoToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(CoredaoToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc **/
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /**
     * Make a query to Coredao explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response Coredao explorer
     */
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].coredaoExplorerApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].coredaoExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    getFullName() {
        return 'Coredao Token';
    }
}
exports.CoredaoToken = CoredaoToken;
CoredaoToken.coinNames = {
    Mainnet: 'coredao',
    Testnet: 'tcoredao',
};
