// @babylonlabs-io/babylon-proto-ts/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.btclightclientquery = exports.btclightclient = exports.incentivequery = exports.incentivetx = exports.epochingquery = exports.epochingtx = exports.btccheckpoint = exports.btcstakingpop = exports.btcstaking = exports.btcstakingtx = void 0;
exports.btcstakingtx = __importStar(require("./generated/babylon/btcstaking/v1/tx"));
exports.btcstaking = __importStar(require("./generated/babylon/btcstaking/v1/btcstaking"));
exports.btcstakingpop = __importStar(require("./generated/babylon/btcstaking/v1/pop"));
exports.btccheckpoint = __importStar(require("./generated/babylon/btccheckpoint/v1/btccheckpoint"));
exports.epochingtx = __importStar(require("./generated/babylon/epoching/v1/tx"));
exports.epochingquery = __importStar(require("./generated/babylon/epoching/v1/query"));
exports.incentivetx = __importStar(require("./generated/babylon/incentive/tx"));
exports.incentivequery = __importStar(require("./generated/babylon/incentive/query"));
exports.btclightclient = __importStar(require("./generated/babylon/btclightclient/v1/btclightclient"));
exports.btclightclientquery = __importStar(require("./generated/babylon/btclightclient/v1/query"));
