// @bitgo/sdk-coin-apt/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BATCH_FUNGIBLE_ASSET_TYPE_TAG = exports.FUNBIGLE_ASSET_TYPE_TAG = exports.DIGITAL_ASSET_TYPE_ARGUMENT = exports.FUNGIBLE_ASSET_TYPE_ARGUMENT = exports.APTOS_COIN = exports.DELEGATION_POOL_WITHDRAW_FUNCTION = exports.DELEGATION_POOL_UNLOCK_FUNCTION = exports.DELEGATION_POOL_ADD_STAKE_FUNCTION = exports.DIGITAL_ASSET_TRANSFER_FUNCTION = exports.COIN_BATCH_TRANSFER_FUNCTION = exports.COIN_TRANSFER_FUNCTION = exports.FUNGIBLE_ASSET_BATCH_TRANSFER_FUNCTION = exports.FUNGIBLE_ASSET_TRANSFER_FUNCTION = exports.DIGITAL_ASSET_TRANSFER_AMOUNT = exports.AMOUNT_BYTES_LENGTH = exports.ADDRESS_BYTES_LENGTH = exports.SECONDS_PER_WEEK = exports.DEFAULT_GAS_UNIT_PRICE = exports.UNAVAILABLE_TEXT = exports.APT_SIGNATURE_LENGTH = exports.APT_BLOCK_ID_LENGTH = exports.APT_TRANSACTION_ID_LENGTH = exports.APT_ADDRESS_LENGTH = void 0;
exports.APT_ADDRESS_LENGTH = 64;
exports.APT_TRANSACTION_ID_LENGTH = 64;
exports.APT_BLOCK_ID_LENGTH = 64;
exports.APT_SIGNATURE_LENGTH = 128;
exports.UNAVAILABLE_TEXT = 'UNAVAILABLE';
exports.DEFAULT_GAS_UNIT_PRICE = 100;
exports.SECONDS_PER_WEEK = 7 * 24 * 60 * 60; // Days * Hours * Minutes * Seconds
exports.ADDRESS_BYTES_LENGTH = 32;
exports.AMOUNT_BYTES_LENGTH = 8;
exports.DIGITAL_ASSET_TRANSFER_AMOUNT = '1';
exports.FUNGIBLE_ASSET_TRANSFER_FUNCTION = '0x1::primary_fungible_store::transfer';
exports.FUNGIBLE_ASSET_BATCH_TRANSFER_FUNCTION = '0x1::aptos_account::batch_transfer_fungible_assets';
exports.COIN_TRANSFER_FUNCTION = '0x1::aptos_account::transfer_coins';
exports.COIN_BATCH_TRANSFER_FUNCTION = '0x1::aptos_account::batch_transfer_coins';
exports.DIGITAL_ASSET_TRANSFER_FUNCTION = '0x1::object::transfer';
exports.DELEGATION_POOL_ADD_STAKE_FUNCTION = '0x1::delegation_pool::add_stake';
exports.DELEGATION_POOL_UNLOCK_FUNCTION = '0x1::delegation_pool::unlock';
exports.DELEGATION_POOL_WITHDRAW_FUNCTION = '0x1::delegation_pool::withdraw';
exports.APTOS_COIN = '0x1::aptos_coin::AptosCoin';
exports.FUNGIBLE_ASSET_TYPE_ARGUMENT = '0x1::fungible_asset::Metadata';
exports.DIGITAL_ASSET_TYPE_ARGUMENT = '0x4::token::Token';
exports.FUNBIGLE_ASSET_TYPE_TAG = '0x1::object::Object';
exports.BATCH_FUNGIBLE_ASSET_TYPE_TAG = '0x1::object::Object<0x1::fungible_asset::Metadata>';
