// @bitgo/sdk-coin-apt/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transferBuilder_1 = require("./transactionBuilder/transferBuilder");
const utils_1 = __importDefault(require("./utils"));
const transaction_1 = require("./transaction/transaction");
const transferTransaction_1 = require("./transaction/transferTransaction");
const fungibleAssetTransfer_1 = require("./transaction/fungibleAssetTransfer");
const fungibleAssetTransferBuilder_1 = require("./transactionBuilder/fungibleAssetTransferBuilder");
const digitalAssetTransfer_1 = require("./transaction/digitalAssetTransfer");
const digitalAssetTransferBuilder_1 = require("./transactionBuilder/digitalAssetTransferBuilder");
const customTransaction_1 = require("./transaction/customTransaction");
const customTransactionBuilder_1 = require("./transactionBuilder/customTransactionBuilder");
const delegationPoolAddStakeTransaction_1 = require("./transaction/delegationPoolAddStakeTransaction");
const delegationPoolAddStakeTransactionBuilder_1 = require("./transactionBuilder/delegationPoolAddStakeTransactionBuilder");
const delegationPoolUnlockTransaction_1 = require("./transaction/delegationPoolUnlockTransaction");
const delegationPoolWithdrawTransactionBuilder_1 = require("./transactionBuilder/delegationPoolWithdrawTransactionBuilder");
const delegationPoolUnlockTransactionBuilder_1 = require("./transactionBuilder/delegationPoolUnlockTransactionBuilder");
const delegationPoolWithdrawTransaction_1 = require("./transaction/delegationPoolWithdrawTransaction");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    from(signedRawTxn, abi) {
        try {
            const signedTxn = transaction_1.Transaction.deserializeSignedTransaction(signedRawTxn);
            const txnType = this.getTransactionTypeFromSignedTxn(signedTxn);
            switch (txnType) {
                case sdk_core_1.TransactionType.Send:
                    const transferTx = new transferTransaction_1.TransferTransaction(this._coinConfig);
                    transferTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getTransferBuilder(transferTx);
                case sdk_core_1.TransactionType.SendToken:
                    const fungibleTransferTokenTx = new fungibleAssetTransfer_1.FungibleAssetTransfer(this._coinConfig);
                    fungibleTransferTokenTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getFungibleAssetTransactionBuilder(fungibleTransferTokenTx);
                case sdk_core_1.TransactionType.SendNFT:
                    const digitalAssetTransferTx = new digitalAssetTransfer_1.DigitalAssetTransfer(this._coinConfig);
                    digitalAssetTransferTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getDigitalAssetTransactionBuilder(digitalAssetTransferTx);
                case sdk_core_1.TransactionType.StakingDelegate:
                    const delegateTx = new delegationPoolAddStakeTransaction_1.DelegationPoolAddStakeTransaction(this._coinConfig);
                    delegateTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getDelegationPoolAddStakeTransactionBuilder(delegateTx);
                case sdk_core_1.TransactionType.StakingUnlock:
                    const unlockTx = new delegationPoolUnlockTransaction_1.DelegationPoolUnlockTransaction(this._coinConfig);
                    unlockTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getDelegationPoolUnlockTransactionBuilder(unlockTx);
                case sdk_core_1.TransactionType.StakingWithdraw:
                    const withdrawTx = new delegationPoolWithdrawTransaction_1.DelegationPoolWithdrawTransaction(this._coinConfig);
                    withdrawTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getDelegationPoolWithdrawTransactionBuilder(withdrawTx);
                case sdk_core_1.TransactionType.CustomTx:
                    const customTx = new customTransaction_1.CustomTransaction(this._coinConfig);
                    if (abi) {
                        customTx.setEntryFunctionAbi(abi);
                    }
                    customTx.fromDeserializedSignedTransaction(signedTxn);
                    return this.getCustomTransactionBuilder(customTx);
                default:
                    throw new sdk_core_1.InvalidTransactionError('Invalid transaction');
            }
        }
        catch (e) {
            throw e;
        }
    }
    getTransactionTypeFromSignedTxn(signedTxn) {
        const rawTxn = signedTxn.raw_txn;
        return utils_1.default.getTransactionTypeFromTransactionPayload(rawTxn.payload);
    }
    /** @inheritdoc */
    getTransferBuilder(tx) {
        return this.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getFungibleAssetTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new fungibleAssetTransferBuilder_1.FungibleAssetTransferBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getDigitalAssetTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new digitalAssetTransferBuilder_1.DigitalAssetTransferBuilder(this._coinConfig));
    }
    getDelegationPoolAddStakeTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new delegationPoolAddStakeTransactionBuilder_1.DelegationPoolAddStakeTransactionBuilder(this._coinConfig));
    }
    getDelegationPoolUnlockTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new delegationPoolUnlockTransactionBuilder_1.DelegationPoolUnlockTransactionBuilder(this._coinConfig));
    }
    getDelegationPoolWithdrawTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new delegationPoolWithdrawTransactionBuilder_1.DelegationPoolWithdrawTransactionBuilder(this._coinConfig));
    }
    /**
     * Get a custom transaction builder
     *
     * @param tx - Optional transaction to initialize the builder with
     * @returns A custom transaction builder
     */
    getCustomTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new customTransactionBuilder_1.CustomTransactionBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getWalletInitializationBuilder() {
        throw new Error('Method not implemented.');
    }
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
