// @bitgo/sdk-coin-apt/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class Utils {
    /** @inheritdoc */
    isValidAddress(address) {
        return this.isValidHex(address, constants_1.APT_ADDRESS_LENGTH);
    }
    /** @inheritdoc */
    isValidBlockId(hash) {
        return this.isValidHex(hash, constants_1.APT_BLOCK_ID_LENGTH);
    }
    /** @inheritdoc */
    isValidPrivateKey(key) {
        return (0, sdk_core_1.isValidEd25519SecretKey)(key);
    }
    /** @inheritdoc */
    isValidPublicKey(key) {
        return (0, sdk_core_1.isValidEd25519PublicKey)(key);
    }
    /** @inheritdoc */
    isValidSignature(signature) {
        return this.isValidHex(signature, constants_1.APT_SIGNATURE_LENGTH);
    }
    /** @inheritdoc */
    isValidTransactionId(txId) {
        return this.isValidHex(txId, constants_1.APT_TRANSACTION_ID_LENGTH);
    }
    isValidHex(value, length) {
        const regex = new RegExp(`^(0x|0X)[a-fA-F0-9]{${length}}$`);
        return regex.test(value);
    }
    getAddressFromPublicKey(publicKey) {
        const aptosPublicKey = new ts_sdk_1.Ed25519PublicKey(Buffer.from(publicKey, 'hex'));
        const authKey = ts_sdk_1.AuthenticationKey.fromPublicKey({ publicKey: aptosPublicKey });
        const accountAddress = authKey.derivedAddress();
        return accountAddress.toString();
    }
    getTransactionTypeFromTransactionPayload(payload) {
        if (!(payload instanceof ts_sdk_1.TransactionPayloadEntryFunction)) {
            throw new Error('Invalid Payload: Expected TransactionPayloadEntryFunction');
        }
        const entryFunction = payload.entryFunction;
        const moduleAddress = entryFunction.module_name.address.toString();
        const moduleIdentifier = entryFunction.module_name.name.identifier;
        const functionIdentifier = entryFunction.function_name.identifier;
        const uniqueIdentifier = `${moduleAddress}::${moduleIdentifier}::${functionIdentifier}`;
        switch (uniqueIdentifier) {
            case constants_1.COIN_TRANSFER_FUNCTION:
            case constants_1.COIN_BATCH_TRANSFER_FUNCTION:
                return sdk_core_1.TransactionType.Send;
            case constants_1.FUNGIBLE_ASSET_TRANSFER_FUNCTION:
            case constants_1.FUNGIBLE_ASSET_BATCH_TRANSFER_FUNCTION:
                return sdk_core_1.TransactionType.SendToken;
            case constants_1.DIGITAL_ASSET_TRANSFER_FUNCTION:
                return sdk_core_1.TransactionType.SendNFT;
            case constants_1.DELEGATION_POOL_ADD_STAKE_FUNCTION:
                return sdk_core_1.TransactionType.StakingDelegate;
            case constants_1.DELEGATION_POOL_UNLOCK_FUNCTION:
                return sdk_core_1.TransactionType.StakingUnlock;
            case constants_1.DELEGATION_POOL_WITHDRAW_FUNCTION:
                return sdk_core_1.TransactionType.StakingWithdraw;
            default:
                // For any other function calls, treat as a custom transaction
                return sdk_core_1.TransactionType.CustomTx;
        }
    }
    fetchAndValidateRecipients(addressArg, amountArg) {
        const addressBytes = addressArg.bcsToBytes();
        const amountBytes = amountArg.bcsToBytes();
        let deserializedAddresses;
        let deserializedAmounts;
        if (addressBytes.length > constants_1.ADDRESS_BYTES_LENGTH || amountBytes.length > constants_1.AMOUNT_BYTES_LENGTH) {
            deserializedAddresses = utils.deserializeAccountAddressVector(addressBytes);
            deserializedAmounts = utils.deserializeU64Vector(amountBytes);
            if (deserializedAddresses.length !== deserializedAmounts.length) {
                console.error('invalid payload entry function arguments : addresses and amounts length mismatch');
                return { recipients: { deserializedAddresses: [], deserializedAmounts: [] }, isValid: false };
            }
        }
        else {
            deserializedAddresses = [addressArg.toString()];
            deserializedAmounts = [amountBytes];
        }
        const allAddressesValid = deserializedAddresses.every((address) => utils.isValidAddress(address.toString()));
        const allAmountsValid = deserializedAmounts.every((amount) => new bignumber_js_1.default(utils.getAmountFromPayloadArgs(amount)).isGreaterThan(0));
        return {
            recipients: { deserializedAddresses, deserializedAmounts },
            isValid: allAddressesValid && allAmountsValid,
        };
    }
    parseRecipients(addressArg, amountArg) {
        const { recipients, isValid } = utils.fetchAndValidateRecipients(addressArg, amountArg);
        if (!isValid) {
            throw new sdk_core_1.InvalidTransactionError('Invalid transaction recipients');
        }
        return recipients.deserializedAddresses.map((address, index) => ({
            address,
            amount: utils.getAmountFromPayloadArgs(recipients.deserializedAmounts[index]),
        }));
    }
    deserializeSignedTransaction(rawTransaction) {
        const txnBytes = ts_sdk_1.Hex.fromHexString(rawTransaction).toUint8Array();
        const deserializer = new ts_sdk_1.Deserializer(txnBytes);
        return deserializer.deserialize(ts_sdk_1.SignedTransaction);
    }
    deserializeFeePayerRawTransaction(rawTransaction) {
        const txnBytes = ts_sdk_1.Hex.fromHexString(rawTransaction).toUint8Array();
        const deserializer = new ts_sdk_1.Deserializer(txnBytes);
        return ts_sdk_1.FeePayerRawTransaction.load(deserializer);
    }
    deserializeAccountAddressVector(serializedBytes) {
        const deserializer = new ts_sdk_1.Deserializer(serializedBytes);
        const deserializedAddresses = deserializer.deserializeVector(ts_sdk_1.AccountAddress);
        return deserializedAddresses.map((address) => address.toString());
    }
    deserializeU64Vector(serializedBytes) {
        const deserializer = new ts_sdk_1.Deserializer(serializedBytes);
        const deserializedAmounts = deserializer.deserializeVector(ts_sdk_1.U64);
        return deserializedAmounts.map((amount) => amount.bcsToBytes());
    }
    getBufferFromHexString(hexString) {
        return Buffer.from(ts_sdk_1.Hex.fromHexString(hexString).toUint8Array());
    }
    castToNumber(value) {
        return new bignumber_js_1.default(value.toString()).toNumber();
    }
    /**
     * Strip hex prefix
     * @param str
     * @returns hex string without 0x prefix
     */
    stripHexPrefix(str) {
        return str.replace(/^0x/i, '');
    }
    getAmountFromPayloadArgs(amountArg) {
        const amountBuffer = Buffer.from(amountArg);
        const low = BigInt(amountBuffer.readUint32LE());
        const high = BigInt(amountBuffer.readUint32LE(4));
        const amount = (high << BigInt(32)) + low;
        return amount.toString();
    }
    /**
     * Returns the Aptos transaction expiration timestamp in seconds.
     * It is set to 1 week from now.
     */
    getTxnExpirationTimestamp() {
        return Math.floor(Date.now() / 1e3) + constants_1.SECONDS_PER_WEEK;
    }
    /**
     * Convert bytes array to hex string with 0x prefix
     */
    bytesToHex(bytes) {
        return '0x' + bytes.map((b) => b.toString(16).padStart(2, '0')).join('');
    }
    /**
     * Try to convert hex string to AccountAddress, return original string if conversion fails
     */
    tryParseAccountAddress(hexStr) {
        try {
            return ts_sdk_1.AccountAddress.fromString(hexStr);
        }
        catch {
            return hexStr;
        }
    }
}
exports.Utils = Utils;
const utils = new Utils();
exports.default = utils;
