// @bitgo/sdk-coin-trx/dist/src/lib/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TronResource = exports.PermissionType = exports.ContractType = void 0;
/**
 * Contract type describes a class of contract in a Transaction. This structure is used to identify what the Contract is.
 */
var ContractType;
(function (ContractType) {
    /**
     * This is a transfer of TRX contract.
     */
    ContractType[ContractType["Transfer"] = 0] = "Transfer";
    /**
     * This is the multi-sig initialization contract.
     */
    ContractType[ContractType["AccountPermissionUpdate"] = 1] = "AccountPermissionUpdate";
    /**
     * This is a smart contract type.
     */
    ContractType[ContractType["TriggerSmartContract"] = 2] = "TriggerSmartContract";
    /**
     * This is the contract for freezeBuilder
     */
    ContractType[ContractType["FreezeBalanceV2"] = 3] = "FreezeBalanceV2";
    /**
     * This is the contract for voting for witnesses
     */
    ContractType[ContractType["VoteWitness"] = 4] = "VoteWitness";
    /**
     * This is the contract for unfreezing balances
     */
    ContractType[ContractType["UnfreezeBalanceV2"] = 5] = "UnfreezeBalanceV2";
    /**
     * This is the contract for withdrawing expired unfrozen balances
     */
    ContractType[ContractType["WithdrawExpireUnfreeze"] = 6] = "WithdrawExpireUnfreeze";
    /**
     * This is the contract for withdrawing reward balances
     */
    ContractType[ContractType["WithdrawBalance"] = 7] = "WithdrawBalance";
    /**
     * This is the contract for delegating resource
     */
    ContractType[ContractType["DelegateResourceContract"] = 8] = "DelegateResourceContract";
    /**
     * This is the contract for un-delegating resource
     */
    ContractType[ContractType["UnDelegateResourceContract"] = 9] = "UnDelegateResourceContract";
})(ContractType || (exports.ContractType = ContractType = {}));
var PermissionType;
(function (PermissionType) {
    PermissionType[PermissionType["Owner"] = 0] = "Owner";
    PermissionType[PermissionType["Witness"] = 1] = "Witness";
    PermissionType[PermissionType["Active"] = 2] = "Active";
})(PermissionType || (exports.PermissionType = PermissionType = {}));
/**
 * Tron resource types
 */
var TronResource;
(function (TronResource) {
    TronResource["BANDWIDTH"] = "BANDWIDTH";
    TronResource["ENERGY"] = "ENERGY";
})(TronResource || (exports.TronResource = TronResource = {}));
