// @bitgo/sdk-coin-trx/dist/src/lib/resourceManagementTxBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceManagementTxBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
const utils_1 = require("./utils");
/**
 * Abstract base class for resource management transaction builders (delegate/undelegate)
 */
class ResourceManagementTxBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._signingKeys = [];
        this.transaction = new transaction_1.Transaction(_coinConfig);
    }
    /**
     * Set the balance amount
     *
     * @param amount amount in TRX to delegate/undelegate
     * @returns the builder with the new parameter set
     */
    setBalance(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._balance = amount;
        return this;
    }
    /**
     * Set the receiver address
     *
     * @param address receiver address for the delegate/undelegate operation
     * @returns the builder with the new parameter set
     */
    setReceiverAddress(address) {
        this.validateAddress(address);
        this._receiverAddress = (0, utils_1.getHexAddressFromBase58Address)(address.address);
        return this;
    }
    /**
     * Set the resource type
     *
     * @param resource resource type to delegate/undelegate
     * @returns the builder with the new parameter set
     */
    setResource(resource) {
        this.validateResource(resource);
        this._resource = resource;
        return this;
    }
    /** @inheritdoc */
    extendValidTo(extensionMs) {
        if (this.transaction.signature && this.transaction.signature.length > 0) {
            throw new sdk_core_1.ExtendTransactionError('Cannot extend a signed transaction');
        }
        if (extensionMs <= 0) {
            throw new Error('Value cannot be below zero');
        }
        if (extensionMs > utils_1.TRANSACTION_MAX_EXPIRATION) {
            throw new sdk_core_1.ExtendTransactionError('The expiration cannot be extended more than one year');
        }
        if (this._expiration) {
            this._expiration = this._expiration + extensionMs;
        }
        else {
            throw new Error('There is not expiration to extend');
        }
    }
    /**
     * Validates the transaction
     *
     * @param {Transaction} transaction - The transaction to validate
     * @throws {void}
     */
    validateTransaction(transaction) {
        this.validateResourceManagementTransactionFields();
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.createResourceManagementTransaction();
        if (this._signingKeys.length > 0) {
            this.applySignatures();
        }
        if (!this.transaction.id) {
            throw new sdk_core_1.BuildTransactionError('A valid transaction must have an id');
        }
        return Promise.resolve(this.transaction);
    }
    /** @inheritdoc */
    signImplementation(key) {
        if (this._signingKeys.some((signingKey) => signingKey.key === key.key)) {
            throw new sdk_core_1.SigningError('Duplicated key');
        }
        this._signingKeys.push(key);
        return this.transaction;
    }
    applySignatures() {
        if (!this.transaction.inputs) {
            throw new sdk_core_1.SigningError('Transaction has no inputs');
        }
        this._signingKeys.forEach((key) => this.applySignature(key));
    }
    /**
     * Validates if the transaction is a valid delegate/undelegate transaction
     *
     * @throws {BuildTransactionError} when the transaction is invalid
     */
    validateResourceManagementTransactionFields() {
        if (!this._balance) {
            throw new sdk_core_1.BuildTransactionError('Missing parameter: balance');
        }
        if (!this._ownerAddress) {
            throw new sdk_core_1.BuildTransactionError('Missing parameter: source');
        }
        if (!this._resource) {
            throw new sdk_core_1.BuildTransactionError('Missing parameter: resource');
        }
        if (!this._receiverAddress) {
            throw new sdk_core_1.BuildTransactionError('Missing parameter: receiver address');
        }
        if (!this._refBlockBytes || !this._refBlockHash) {
            throw new sdk_core_1.BuildTransactionError('Missing block reference information');
        }
    }
    /**
     * Initialize the delegate/undelegate contract call specific data
     *
     * @param {ResourceManagementContract} resourceManagementContractCall object with delegate txn data
     */
    initResourceManagementContractCall(resourceManagementContractCall) {
        const { resource, owner_address, balance, receiver_address } = resourceManagementContractCall.parameter.value;
        if (owner_address) {
            this.source({ address: (0, utils_1.getBase58AddressFromHex)(owner_address) });
        }
        if (resource) {
            this.setResource(resource);
        }
        if (balance !== undefined) {
            this.setBalance(balance.toString());
        }
        if (receiver_address) {
            this.setReceiverAddress({ address: (0, utils_1.getBase58AddressFromHex)(receiver_address) });
        }
    }
}
exports.ResourceManagementTxBuilder = ResourceManagementTxBuilder;
