// @bitgo/sdk-coin-trx/dist/src/lib/undelegateResourceTxBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UndelegateResourceTxBuilder = void 0;
const crypto_1 = require("crypto");
const sdk_core_1 = require("@bitgo/sdk-core");
const transaction_1 = require("./transaction");
const constants_1 = require("./constants");
const utils_1 = require("./utils");
const tron_1 = require("../../resources/protobuf/tron");
const resourceManagementTxBuilder_1 = require("./resourceManagementTxBuilder");
var ContractType = tron_1.protocol.Transaction.Contract.ContractType;
class UndelegateResourceTxBuilder extends resourceManagementTxBuilder_1.ResourceManagementTxBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    get transactionType() {
        return sdk_core_1.TransactionType.UnDelegateResource;
    }
    /**
     * Initialize the transaction builder fields using the transaction data
     *
     * @param {TransactionReceipt | string} rawTransaction the transaction data in a string or JSON format
     * @returns {DelegateResourceTxBuilder} the builder with the transaction data set
     */
    initBuilder(rawTransaction) {
        this.validateRawTransaction(rawTransaction);
        const tx = this.fromImplementation(rawTransaction);
        this.transaction = tx;
        this._signingKeys = [];
        const rawData = tx.toJson().raw_data;
        this._refBlockBytes = rawData.ref_block_bytes;
        this._refBlockHash = rawData.ref_block_hash;
        this._expiration = rawData.expiration;
        this._timestamp = rawData.timestamp;
        this.transaction.setTransactionType(sdk_core_1.TransactionType.UnDelegateResource);
        const contractCall = rawData.contract[0];
        this.initResourceManagementContractCall(contractCall);
        return this;
    }
    /**
     * Helper method to create the undelegate resource transaction
     */
    createResourceManagementTransaction() {
        const rawDataHex = this.getResourceManagementTxRawDataHex();
        const rawData = (0, utils_1.decodeTransaction)(rawDataHex);
        const contract = rawData.contract[0];
        const contractParameter = contract.parameter;
        contractParameter.value.owner_address = this._ownerAddress.toLocaleLowerCase();
        contractParameter.value.balance = Number(this._balance);
        contractParameter.value.receiver_address = this._receiverAddress.toLocaleLowerCase();
        contractParameter.value.resource = this._resource;
        contractParameter.type_url = constants_1.UNDELEGATION_TYPE_URL;
        contract.type = 'UnDelegateResourceContract';
        const hexBuffer = Buffer.from(rawDataHex, 'hex');
        const id = (0, crypto_1.createHash)('sha256').update(hexBuffer).digest('hex');
        const txReceipt = {
            raw_data: rawData,
            raw_data_hex: rawDataHex,
            txID: id,
            signature: this.transaction.signature,
        };
        this.transaction = new transaction_1.Transaction(this._coinConfig, txReceipt);
    }
    /**
     * Helper method to get the undelegate resource transaction raw data hex
     *
     * @returns {string} the undelegate resource transaction raw data hex
     */
    getResourceManagementTxRawDataHex() {
        const rawContract = {
            ownerAddress: (0, utils_1.getByteArrayFromHexAddress)(this._ownerAddress),
            receiverAddress: (0, utils_1.getByteArrayFromHexAddress)(this._receiverAddress),
            balance: this._balance,
        };
        // Only include resource if it's not BANDWIDTH (the default value = 0)
        // In protobuf3, default values are typically not encoded in the wire format.
        // TRON's node re-serializes transactions and omits default values,
        // so we must match that behavior to ensure consistent transaction hashes.
        if (this._resource !== 'BANDWIDTH') {
            rawContract.resource = this._resource;
        }
        const undelegateResourceContract = tron_1.protocol.UnDelegateResourceContract.fromObject(rawContract);
        const undelegateResourceContractBytes = tron_1.protocol.UnDelegateResourceContract.encode(undelegateResourceContract).finish();
        const txContract = {
            type: ContractType.UnDelegateResourceContract,
            parameter: {
                value: undelegateResourceContractBytes,
                type_url: constants_1.UNDELEGATION_TYPE_URL,
            },
        };
        const raw = {
            refBlockBytes: Buffer.from(this._refBlockBytes, 'hex'),
            refBlockHash: Buffer.from(this._refBlockHash, 'hex'),
            expiration: this._expiration || Date.now() + utils_1.TRANSACTION_DEFAULT_EXPIRATION,
            timestamp: this._timestamp || Date.now(),
            contract: [txContract],
        };
        const rawTx = tron_1.protocol.Transaction.raw.create(raw);
        return Buffer.from(tron_1.protocol.Transaction.raw.encode(rawTx).finish()).toString('hex');
    }
}
exports.UndelegateResourceTxBuilder = UndelegateResourceTxBuilder;
