// @bitgo/sdk-coin-trx/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALID_RESOURCE_TYPES = exports.tokenTestnetContractAddresses = exports.tokenMainnetContractAddresses = exports.TRANSACTION_DEFAULT_EXPIRATION = exports.TRANSACTION_MAX_EXPIRATION = void 0;
exports.isBase58Address = isBase58Address;
exports.getByteArrayFromHexAddress = getByteArrayFromHexAddress;
exports.getHexAddressFromByteArray = getHexAddressFromByteArray;
exports.verifySignature = verifySignature;
exports.getHexAddressFromBase58Address = getHexAddressFromBase58Address;
exports.getPubKeyFromPriKey = getPubKeyFromPriKey;
exports.getAddressFromPriKey = getAddressFromPriKey;
exports.getBase58AddressFromByteArray = getBase58AddressFromByteArray;
exports.getBase58AddressFromHex = getBase58AddressFromHex;
exports.signTransaction = signTransaction;
exports.signString = signString;
exports.getRawAddressFromPubKey = getRawAddressFromPubKey;
exports.decodeTransaction = decodeTransaction;
exports.decodeRawTransaction = decodeRawTransaction;
exports.getHexFromBase64 = getHexFromBase64;
exports.isValidHex = isValidHex;
exports.decodeTransferContract = decodeTransferContract;
exports.decodeTriggerSmartContract = decodeTriggerSmartContract;
exports.decodeAccountPermissionUpdateContract = decodeAccountPermissionUpdateContract;
exports.decodeFreezeBalanceV2Contract = decodeFreezeBalanceV2Contract;
exports.decodeVoteWitnessContract = decodeVoteWitnessContract;
exports.decodeUnfreezeBalanceV2Contract = decodeUnfreezeBalanceV2Contract;
exports.decodeWithdrawBalanceContract = decodeWithdrawBalanceContract;
exports.decodeWithdrawExpireUnfreezeContract = decodeWithdrawExpireUnfreezeContract;
exports.decodeDelegateResourceContract = decodeDelegateResourceContract;
exports.decodeUnDelegateResourceContract = decodeUnDelegateResourceContract;
exports.isValidTxJsonString = isValidTxJsonString;
exports.isValidRawTransactionFormat = isValidRawTransactionFormat;
exports.toHex = toHex;
exports.encodeDataParams = encodeDataParams;
exports.decodeDataParams = decodeDataParams;
exports.generateRawDataHex = generateRawDataHex;
const assert_1 = __importDefault(require("assert"));
const hex = __importStar(require("@stablelib/hex"));
const tronweb = __importStar(require("tronweb"));
const tron_1 = require("../../resources/protobuf/tron");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const enum_1 = require("./enum");
const utils_1 = require("ethers/lib/utils");
const constants_1 = require("./constants");
exports.TRANSACTION_MAX_EXPIRATION = 86400000; // one day
exports.TRANSACTION_DEFAULT_EXPIRATION = 10800000; // three hours
const ADDRESS_PREFIX_REGEX = /^(41)/;
const ADDRESS_PREFIX = '41';
const getTronTokens = (network = 'mainnet') => {
    return statics_1.coins
        .filter((coin) => coin.family === 'trx')
        .filter((trx) => trx.network.type === network && trx.isToken).map((coins) => coins.contractAddress.toString());
};
exports.tokenMainnetContractAddresses = getTronTokens('mainnet');
exports.tokenTestnetContractAddresses = getTronTokens('testnet');
exports.VALID_RESOURCE_TYPES = ['ENERGY', 'BANDWIDTH'];
/**
 * @param address
 */
function isBase58Address(address) {
    return tronweb.utils.crypto.isAddressValid(address);
}
/**
 * @param str
 */
function getByteArrayFromHexAddress(str) {
    return tronweb.utils.code.hexStr2byteArray(str.replace('0x', ''));
}
/**
 * @param arr
 */
function getHexAddressFromByteArray(arr) {
    return tronweb.utils.code.byteArray2hexStr(arr);
}
/**
 * @param messageToVerify
 * @param base58Address
 * @param sigHex
 * @param useTronHeader
 */
function verifySignature(messageToVerify, base58Address, sigHex, useTronHeader = true) {
    if (!isValidHex(sigHex)) {
        throw new sdk_core_1.UtilsError('signature is not in a valid format, needs to be hexadecimal');
    }
    if (!isValidHex(messageToVerify)) {
        throw new sdk_core_1.UtilsError('message is not in a valid format, needs to be hexadecimal');
    }
    if (!isBase58Address(base58Address)) {
        throw new sdk_core_1.UtilsError('address needs to be base58 encoded');
    }
    return tronweb.Trx.verifySignature(messageToVerify, base58Address, sigHex, useTronHeader);
}
/**
 * @param base58
 */
function getHexAddressFromBase58Address(base58) {
    // pulled from: https://github.com/TRON-US/tronweb/blob/dcb8efa36a5ebb65c4dab3626e90256a453f3b0d/src/utils/help.js#L17
    // but they don't surface this call in index.js
    const bytes = tronweb.utils.crypto.decodeBase58Address(base58);
    // Ensure bytes is a ByteArray (number[])
    if (Array.isArray(bytes)) {
        return getHexAddressFromByteArray(bytes);
    }
    throw new sdk_core_1.UtilsError('Failed to decode base58 address to byte array');
}
/**
 * @param privateKey
 */
function getPubKeyFromPriKey(privateKey) {
    return tronweb.utils.crypto.getPubKeyFromPriKey(privateKey);
}
/**
 * @param privateKey
 */
function getAddressFromPriKey(privateKey) {
    return tronweb.utils.crypto.getAddressFromPriKey(privateKey);
}
/**
 * @param address
 */
function getBase58AddressFromByteArray(address) {
    return tronweb.utils.crypto.getBase58CheckAddress(address);
}
/**
 * @param hex
 */
function getBase58AddressFromHex(hex) {
    const arr = getByteArrayFromHexAddress(hex);
    return getBase58AddressFromByteArray(arr);
}
/**
 * @param privateKey
 * @param transaction
 */
function signTransaction(privateKey, transaction) {
    return tronweb.utils.crypto.signTransaction(privateKey, transaction);
}
/**
 * @param message
 * @param privateKey
 * @param useTronHeader
 */
function signString(message, privateKey, useTronHeader = true) {
    return tronweb.Trx.signString(message, privateKey, useTronHeader);
}
/**
 * @param pubBytes
 */
function getRawAddressFromPubKey(pubBytes) {
    return tronweb.utils.crypto.computeAddress(pubBytes);
}
/**
 * Decodes a hex encoded transaction in its protobuf representation.
 *
 * @param hexString raw_data_hex field from tron transactions
 */
function decodeTransaction(hexString) {
    const rawTransaction = decodeRawTransaction(hexString);
    // there should not be multiple contracts in this data
    if (rawTransaction.contracts.length !== 1) {
        throw new sdk_core_1.UtilsError('Number of contracts is greater than 1.');
    }
    let contract;
    let contractType;
    // ensure the contract type is supported
    switch (rawTransaction.contracts[0].parameter.type_url) {
        case 'type.googleapis.com/protocol.TransferContract':
            contractType = enum_1.ContractType.Transfer;
            contract = exports.decodeTransferContract(rawTransaction.contracts[0].parameter.value);
            break;
        case 'type.googleapis.com/protocol.AccountPermissionUpdateContract':
            contractType = enum_1.ContractType.AccountPermissionUpdate;
            contract = exports.decodeAccountPermissionUpdateContract(rawTransaction.contracts[0].parameter.value);
            break;
        case 'type.googleapis.com/protocol.TriggerSmartContract':
            contractType = enum_1.ContractType.TriggerSmartContract;
            contract = exports.decodeTriggerSmartContract(rawTransaction.contracts[0].parameter.value);
            break;
        case 'type.googleapis.com/protocol.FreezeBalanceV2Contract':
            contractType = enum_1.ContractType.FreezeBalanceV2;
            contract = decodeFreezeBalanceV2Contract(rawTransaction.contracts[0].parameter.value);
            break;
        case 'type.googleapis.com/protocol.VoteWitnessContract':
            contractType = enum_1.ContractType.VoteWitness;
            contract = decodeVoteWitnessContract(rawTransaction.contracts[0].parameter.value);
            break;
        case 'type.googleapis.com/protocol.WithdrawExpireUnfreezeContract':
            contract = decodeWithdrawExpireUnfreezeContract(rawTransaction.contracts[0].parameter.value);
            contractType = enum_1.ContractType.WithdrawExpireUnfreeze;
            break;
        case 'type.googleapis.com/protocol.WithdrawBalanceContract':
            contract = decodeWithdrawBalanceContract(rawTransaction.contracts[0].parameter.value);
            contractType = enum_1.ContractType.WithdrawBalance;
            break;
        case 'type.googleapis.com/protocol.UnfreezeBalanceV2Contract':
            contract = decodeUnfreezeBalanceV2Contract(rawTransaction.contracts[0].parameter.value);
            contractType = enum_1.ContractType.UnfreezeBalanceV2;
            break;
        case constants_1.DELEGATION_TYPE_URL:
            contractType = enum_1.ContractType.DelegateResourceContract;
            contract = decodeDelegateResourceContract(rawTransaction.contracts[0].parameter.value);
            break;
        case 'type.googleapis.com/protocol.UnDelegateResourceContract':
            contractType = enum_1.ContractType.UnDelegateResourceContract;
            contract = decodeUnDelegateResourceContract(rawTransaction.contracts[0].parameter.value);
            break;
        default:
            throw new sdk_core_1.UtilsError('Unsupported contract type');
    }
    return {
        contractType,
        contract,
        expiration: rawTransaction.expiration,
        timestamp: rawTransaction.timestamp,
        ref_block_bytes: rawTransaction.blockBytes,
        ref_block_hash: rawTransaction.blockHash,
        fee_limit: +rawTransaction.feeLimit,
    };
}
/**
 * Decodes a transaction's raw field from a base64 encoded string. This is a protobuf representation.
 *
 * @param hexString this is the raw hexadecimal encoded string. Doc found in the following link.
 * @example
 * @see {@link https://github.com/BitGo/bitgo-account-lib/blob/5f282588701778a4421c75fa61f42713f56e95b9/resources/protobuf/tron.proto#L319}
 */
function decodeRawTransaction(hexString) {
    const bytes = Buffer.from(hexString, 'hex');
    let raw;
    try {
        // we need to decode our raw_data_hex field first
        raw = tron_1.protocol.Transaction.raw.decode(bytes);
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the initial raw_data_hex from the serialized tx.');
    }
    return {
        expiration: Number(raw.expiration),
        timestamp: Number(raw.timestamp),
        contracts: raw.contract,
        blockBytes: toHex(raw.refBlockBytes),
        feeLimit: raw.feeLimit,
        blockHash: toHex(raw.refBlockHash),
    };
}
/**
 * Converts a base64 encoded string to hex
 *
 * @param base64 - The base64 encoded string to convert
 * @returns {string} - The hex representation
 */
function getHexFromBase64(base64) {
    const buffer = Buffer.from(base64, 'base64');
    return buffer.toString('hex');
}
/**
 * Indicates whether the passed string is a safe hex string for tron's purposes.
 *
 * @param hex A valid hex string must be a string made of numbers and characters and has an even length.
 */
function isValidHex(hex) {
    return /^(0x)?([0-9a-f]{2})+$/i.test(hex);
}
/** Deserialize the segment of the txHex which corresponds with the details of the transfer
 *
 * @param transferHex is the value property of the "parameter" field of contractList[0]
 * */
function decodeTransferContract(transferHex) {
    const contractBytes = Buffer.from(transferHex, 'base64');
    let transferContract;
    try {
        transferContract = tron_1.protocol.TransferContract.decode(contractBytes);
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the transfer contract in the transaction.');
    }
    if (!transferContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this transfer contract.');
    }
    if (!transferContract.toAddress) {
        throw new sdk_core_1.UtilsError('Destination address does not exist in this transfer contract.');
    }
    if (!transferContract.hasOwnProperty('amount')) {
        throw new sdk_core_1.UtilsError('Amount does not exist in this transfer contract.');
    }
    // deserialize attributes
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(transferContract.ownerAddress, 'base64').toString('hex')));
    const to_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(transferContract.toAddress, 'base64').toString('hex')));
    const amount = transferContract.amount;
    return [
        {
            parameter: {
                value: {
                    amount: Number(amount),
                    owner_address,
                    to_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with trigger smart contract
 *
 * @param {string} base64
 * @returns {AccountPermissionUpdateContract}
 */
function decodeTriggerSmartContract(base64) {
    let contractCallDecoded;
    try {
        contractCallDecoded = tron_1.protocol.TriggerSmartContract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the contract call in the transaction.');
    }
    if (!contractCallDecoded.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this contract call.');
    }
    if (!contractCallDecoded.contractAddress) {
        throw new sdk_core_1.UtilsError('Destination contract address does not exist in this contract call.');
    }
    if (!contractCallDecoded.data) {
        throw new sdk_core_1.UtilsError('Data does not exist in this contract call.');
    }
    // deserialize attributes
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(contractCallDecoded.ownerAddress, 'base64').toString('hex')));
    const contract_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(contractCallDecoded.contractAddress, 'base64').toString('hex')));
    const data = contractCallDecoded.data;
    return [
        {
            parameter: {
                value: {
                    data: data,
                    owner_address,
                    contract_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with the details of the contract which updates
 * account permission
 *
 * @param {string} base64
 * @returns {AccountPermissionUpdateContract}
 */
function decodeAccountPermissionUpdateContract(base64) {
    const accountUpdateContract = tron_1.protocol.AccountPermissionUpdateContract.decode(Buffer.from(base64, 'base64')).toJSON();
    (0, assert_1.default)(accountUpdateContract.ownerAddress);
    (0, assert_1.default)(accountUpdateContract.owner);
    (0, assert_1.default)(accountUpdateContract.hasOwnProperty('actives'));
    const ownerAddress = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(accountUpdateContract.ownerAddress, 'base64').toString('hex')));
    const owner = createPermission(accountUpdateContract.owner);
    let witness = undefined;
    if (accountUpdateContract.witness) {
        witness = createPermission(accountUpdateContract.witness);
    }
    const activeList = accountUpdateContract.actives.map((active) => createPermission(active));
    return {
        ownerAddress,
        owner,
        witness,
        actives: activeList,
    };
}
/**
 * Deserialize the segment of the txHex corresponding with freeze balance contract
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {FreezeBalanceContractParameter[]} - Array containing the decoded freeze contract
 */
function decodeFreezeBalanceV2Contract(base64) {
    let freezeContract;
    try {
        freezeContract = tron_1.protocol.FreezeBalanceV2Contract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the freeze contract in the transaction.');
    }
    if (!freezeContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this freeze contract.');
    }
    if (freezeContract.frozenBalance === undefined) {
        throw new sdk_core_1.UtilsError('Frozen balance does not exist in this freeze contract.');
    }
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(freezeContract.ownerAddress, 'base64').toString('hex')));
    // In protobuf3, default values (BANDWIDTH = 0) may be omitted from serialization.
    // If resource is undefined, default to BANDWIDTH.
    const resourceValue = freezeContract.resource === undefined || freezeContract.resource === 'BANDWIDTH'
        ? enum_1.TronResource.BANDWIDTH
        : enum_1.TronResource.ENERGY;
    return [
        {
            parameter: {
                value: {
                    resource: resourceValue,
                    frozen_balance: Number(freezeContract.frozenBalance),
                    owner_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with vote witness contract
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {VoteWitnessContractParameter[]} - Array containing the decoded vote witness contract
 */
function decodeVoteWitnessContract(base64) {
    let voteContract;
    try {
        voteContract = tron_1.protocol.VoteWitnessContract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the vote contract in the transaction.');
    }
    if (!voteContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this vote contract.');
    }
    if (!Array.isArray(voteContract.votes) || voteContract.votes.length === 0) {
        throw new sdk_core_1.UtilsError('Votes do not exist or are empty in this vote contract.');
    }
    // deserialize attributes
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(voteContract.ownerAddress, 'base64').toString('hex')));
    const votes = voteContract.votes.map((vote) => {
        if (!vote.voteAddress) {
            throw new sdk_core_1.UtilsError('Vote address is missing in one of the votes.');
        }
        return {
            vote_address: getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(vote.voteAddress, 'base64').toString('hex'))),
            vote_count: Number(vote.voteCount || 0),
        };
    });
    return [
        {
            parameter: {
                value: {
                    owner_address,
                    votes,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with unfreeze balance contract
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {UnfreezeBalanceContractParameter[]} - Array containing the decoded unfreeze contract
 */
function decodeUnfreezeBalanceV2Contract(base64) {
    let unfreezeContract;
    try {
        unfreezeContract = tron_1.protocol.UnfreezeBalanceV2Contract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the unfreeze contract in the transaction.');
    }
    if (!unfreezeContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this unfreeze contract.');
    }
    if (unfreezeContract.unfreezeBalance === undefined) {
        throw new sdk_core_1.UtilsError('Unfreeze balance does not exist in this unfreeze contract.');
    }
    // deserialize attributes
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(unfreezeContract.ownerAddress, 'base64').toString('hex')));
    // In protobuf3, default values (BANDWIDTH = 0) may be omitted from serialization.
    // If resource is undefined, default to BANDWIDTH.
    const resourceValue = unfreezeContract.resource;
    const resourceEnum = resourceValue === undefined || resourceValue === tron_1.protocol.ResourceCode.BANDWIDTH
        ? enum_1.TronResource.BANDWIDTH
        : enum_1.TronResource.ENERGY;
    return [
        {
            parameter: {
                value: {
                    resource: resourceEnum,
                    unfreeze_balance: Number(unfreezeContract.unfreezeBalance),
                    owner_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with withdraw balance contract
 * Decoded contract is the same as withdraw expire unfreeze
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {WithdrawExpireUnfreezeContractParameter[]} - Array containing the decoded withdraw contract
 */
function decodeWithdrawBalanceContract(base64) {
    let withdrawContract;
    try {
        withdrawContract = tron_1.protocol.WithdrawBalanceContract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the withdraw contract in the transaction.');
    }
    if (!withdrawContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this withdraw contract.');
    }
    // deserialize attributes
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(withdrawContract.ownerAddress, 'base64').toString('hex')));
    return [
        {
            parameter: {
                value: {
                    owner_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with withdraw expire unfreeze contract
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {WithdrawExpireUnfreezeContractParameter[]} - Array containing the decoded withdraw contract
 */
function decodeWithdrawExpireUnfreezeContract(base64) {
    let withdrawContract;
    try {
        withdrawContract = tron_1.protocol.WithdrawBalanceContract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the withdraw contract in the transaction.');
    }
    if (!withdrawContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this withdraw contract.');
    }
    // deserialize attributes
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(withdrawContract.ownerAddress, 'base64').toString('hex')));
    return [
        {
            parameter: {
                value: {
                    owner_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with delegate resource contract
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {ResourceManagementContractParameter[]} - Array containing the decoded delegate resource contract
 */
function decodeDelegateResourceContract(base64) {
    let delegateResourceContract;
    try {
        delegateResourceContract = tron_1.protocol.DelegateResourceContract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the delegate resource contract in the transaction.');
    }
    if (!delegateResourceContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this delegate resource contract.');
    }
    if (!delegateResourceContract.receiverAddress) {
        throw new sdk_core_1.UtilsError('Receiver address does not exist in this delegate resource contract.');
    }
    if (delegateResourceContract.balance === undefined) {
        throw new sdk_core_1.UtilsError('Balance does not exist in this delegate resource contract.');
    }
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(delegateResourceContract.ownerAddress, 'base64').toString('hex')));
    const receiver_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(delegateResourceContract.receiverAddress, 'base64').toString('hex')));
    // In protobuf3, default values (BANDWIDTH = 0) may be omitted from serialization.
    // If resource is undefined or falsy, default to BANDWIDTH.
    const resourceValue = !delegateResourceContract.resource ? enum_1.TronResource.BANDWIDTH : enum_1.TronResource.ENERGY;
    return [
        {
            parameter: {
                value: {
                    resource: resourceValue,
                    balance: Number(delegateResourceContract.balance),
                    owner_address,
                    receiver_address,
                },
            },
        },
    ];
}
/**
 * Deserialize the segment of the txHex corresponding with undelegate resource contract
 *
 * @param {string} base64 - The base64 encoded contract data
 * @returns {ResourceManagementContractParameter[]} - Array containing the decoded undelegate resource contract
 */
function decodeUnDelegateResourceContract(base64) {
    let undelegateResourceContract;
    try {
        undelegateResourceContract = tron_1.protocol.UnDelegateResourceContract.decode(Buffer.from(base64, 'base64')).toJSON();
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('There was an error decoding the delegate resource contract in the transaction.');
    }
    if (!undelegateResourceContract.ownerAddress) {
        throw new sdk_core_1.UtilsError('Owner address does not exist in this delegate resource contract.');
    }
    if (!undelegateResourceContract.receiverAddress) {
        throw new sdk_core_1.UtilsError('Receiver address does not exist in this delegate resource contract.');
    }
    if (undelegateResourceContract.balance === undefined) {
        throw new sdk_core_1.UtilsError('Balance does not exist in this delegate resource contract.');
    }
    const owner_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(undelegateResourceContract.ownerAddress, 'base64').toString('hex')));
    const receiver_address = getBase58AddressFromByteArray(getByteArrayFromHexAddress(Buffer.from(undelegateResourceContract.receiverAddress, 'base64').toString('hex')));
    // In protobuf3, default values (BANDWIDTH = 0) may be omitted from serialization.
    // If resource is undefined or falsy, default to BANDWIDTH.
    const resourceValue = !undelegateResourceContract.resource ? enum_1.TronResource.BANDWIDTH : enum_1.TronResource.ENERGY;
    return [
        {
            parameter: {
                value: {
                    resource: resourceValue,
                    balance: Number(undelegateResourceContract.balance),
                    owner_address,
                    receiver_address,
                },
            },
        },
    ];
}
/**
 * @param raw
 */
function createPermission(raw) {
    let permissionType;
    const permission = raw.permissionName.toLowerCase().trim();
    if (permission === 'owner') {
        permissionType = enum_1.PermissionType.Owner;
    }
    else if (permission === 'witness') {
        permissionType = enum_1.PermissionType.Witness;
    }
    else if (permission.substr(0, 6) === 'active') {
        permissionType = enum_1.PermissionType.Active;
    }
    else {
        throw new sdk_core_1.UtilsError('Permission type not parseable.');
    }
    return { type: permissionType, threshold: raw.threshold };
}
/**
 * @param rawTransaction
 */
function isValidTxJsonString(rawTransaction) {
    const transaction = JSON.parse(rawTransaction);
    return transaction.hasOwnProperty('txID');
}
/**
 * Returns whether the provided raw transaction accommodates to bitgo's preferred format
 *
 * @param {any} rawTransaction - The raw transaction to be checked
 * @returns {boolean} the validation result
 */
function isValidRawTransactionFormat(rawTransaction) {
    if (typeof rawTransaction === 'string' && (isValidHex(rawTransaction) || isValidTxJsonString(rawTransaction))) {
        return true;
    }
    return false;
}
/**
 * Returns an hex string of the given buffer
 *
 * @param {Buffer | Uint8Array} buffer - the buffer to be converted to hex
 * @returns {string} - the hex value
 */
function toHex(buffer) {
    return hex.encode(buffer, true);
}
/**
 * Returns a Keccak-256 encoded string of the parameters
 *
 * @param types - strings describing the types of the values
 * @param values - value to encode
 * @param methodId - the first 4 bytes of the function selector
 */
function encodeDataParams(types, values, methodId) {
    types.forEach((type, index) => {
        if (type == 'address') {
            values[index] = values[index].replace(ADDRESS_PREFIX_REGEX, '0x');
        }
    });
    const abiCoder = new utils_1.AbiCoder();
    let data;
    try {
        data = abiCoder.encode(types, values);
    }
    catch (e) {
        throw new sdk_core_1.UtilsError(`There was an error encoding the data params. Error = ${JSON.stringify(e)}`);
    }
    if (methodId) {
        return (0, utils_1.hexConcat)([methodId, data]).replace(/^(0x)/, '');
    }
    else {
        return data.replace(/^(0x)/, '');
    }
}
/**
 * Returns the decoded values according to the array of types
 *
 * @param types - strings describing the types of the values
 * @param data - encoded string
 */
function decodeDataParams(types, data) {
    const abiCoder = new utils_1.AbiCoder();
    data = '0x' + data.substring(8);
    return abiCoder.decode(types, data).reduce((obj, arg, index) => {
        if (types[index] == 'address')
            arg = ADDRESS_PREFIX + arg.substr(2).toLowerCase();
        obj.push(arg);
        return obj;
    }, []);
}
/**
 * Generate raw_data_hex for a TRON transaction
 *
 * @param {Object} rawData - The transaction raw data object containing:
 * @param {Array} rawData.contract - Array of contract objects
 * @param {string} rawData.refBlockBytes - Reference block bytes
 * @param {string} rawData.refBlockHash - Reference block hash
 * @param {number} rawData.expiration - Transaction expiration timestamp
 * @param {number} rawData.timestamp - Transaction creation timestamp
 * @param {number} [rawData.feeLimit] - Optional fee limit for smart contracts
 * @returns {string} The hex string representation of the encoded transaction data
 */
function generateRawDataHex(rawData = {}) {
    try {
        // Process contracts to ensure proper protobuf encoding
        let processedContracts = rawData.contract;
        if (rawData.contract && rawData.contract.length > 0) {
            processedContracts = rawData.contract.map((contract) => {
                // Handle TransferContract specifically
                if (contract.parameter?.type_url === 'type.googleapis.com/protocol.TransferContract') {
                    const contractValue = contract.parameter.value;
                    // Create the protobuf contract object
                    const transferContract = {};
                    // Handle owner_address (required field)
                    if (contractValue.owner_address) {
                        transferContract.ownerAddress = Buffer.from(contractValue.owner_address, 'hex');
                    }
                    // Handle to_address (required field)
                    if (contractValue.to_address) {
                        transferContract.toAddress = Buffer.from(contractValue.to_address, 'hex');
                    }
                    // Handle amount (required field)
                    if (contractValue.amount !== undefined) {
                        transferContract.amount = contractValue.amount;
                    }
                    // Encode the contract using protobuf
                    const encodedContract = tron_1.protocol.TransferContract.encode(transferContract).finish();
                    const base64Value = Buffer.from(encodedContract).toString('base64');
                    return {
                        ...contract,
                        parameter: {
                            ...contract.parameter,
                            value: base64Value,
                        },
                    };
                }
                return contract;
            });
        }
        // Create raw transaction object matching protobuf schema
        const rawTx = {
            contract: processedContracts,
            refBlockBytes: rawData.refBlockBytes ? Buffer.from(rawData.refBlockBytes, 'hex') : undefined,
            refBlockHash: rawData.refBlockHash ? Buffer.from(rawData.refBlockHash, 'hex') : undefined,
            expiration: rawData.expiration,
            timestamp: rawData.timestamp,
            feeLimit: rawData.feeLimit,
        };
        // Encode using protobuf and get final bytes
        const encodedBytes = tron_1.protocol.Transaction.raw.encode(rawTx).finish();
        // Convert to hex string
        return Buffer.from(encodedBytes).toString('hex');
    }
    catch (e) {
        throw new sdk_core_1.UtilsError('Failed to generate raw data hex: ' + e.message);
    }
}
