// @bitgo/sdk-coin-trx/dist/src/lib/voteWitnessTxBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoteWitnessTxBuilder = void 0;
const crypto_1 = require("crypto");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
const utils_1 = require("./utils");
const tron_1 = require("../../resources/protobuf/tron");
var ContractType = tron_1.protocol.Transaction.Contract.ContractType;
class VoteWitnessTxBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._signingKeys = [];
        this.transaction = new transaction_1.Transaction(_coinConfig);
    }
    /** @inheritdoc */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingVote;
    }
    /**
     * Set the votes to be used in the transaction
     *
     * @param votes vote data containing vote address and vote count
     * @returns the builder with the new parameter set
     */
    setVotes(votes) {
        this.validateVotes(votes);
        this._votes = votes.map((vote) => ({
            ...vote,
            vote_address: (0, utils_1.getHexAddressFromBase58Address)(vote.vote_address),
        }));
        return this;
    }
    /** @inheritdoc */
    extendValidTo(extensionMs) {
        if (this.transaction.signature && this.transaction.signature.length > 0) {
            throw new sdk_core_1.ExtendTransactionError('Cannot extend a signed transaction');
        }
        if (extensionMs <= 0) {
            throw new Error('Value cannot be below zero');
        }
        if (extensionMs > utils_1.TRANSACTION_MAX_EXPIRATION) {
            throw new sdk_core_1.ExtendTransactionError('The expiration cannot be extended more than one year');
        }
        if (this._expiration) {
            this._expiration = this._expiration + extensionMs;
        }
        else {
            throw new Error('There is not expiration to extend');
        }
    }
    /**
     * Initialize the transaction builder fields using the transaction data
     *
     * @param {TransactionReceipt | string} rawTransaction the transaction data in a string or JSON format
     * @returns {VoteWitnessTxBuilder} the builder with the transaction data set
     */
    initBuilder(rawTransaction) {
        this.transaction = this.fromImplementation(rawTransaction);
        this.transaction.setTransactionType(this.transactionType);
        this.validateRawTransaction(rawTransaction);
        const tx = this.fromImplementation(rawTransaction);
        this.transaction = tx;
        this._signingKeys = [];
        const rawData = tx.toJson().raw_data;
        this._refBlockBytes = rawData.ref_block_bytes;
        this._refBlockHash = rawData.ref_block_hash;
        this._expiration = rawData.expiration;
        this._timestamp = rawData.timestamp;
        this.transaction.setTransactionType(sdk_core_1.TransactionType.StakingVote);
        const contractCall = rawData.contract[0];
        this.initVoteWitnessContractCall(contractCall);
        return this;
    }
    /**
     * Initialize the votewitnesscontract call specific data
     *
     * @param {VoteWitnessContract} contractCall object with transfer data
     */
    initVoteWitnessContractCall(voteWitnessContractCall) {
        const { votes, owner_address } = voteWitnessContractCall.parameter.value;
        if (owner_address) {
            this.source({ address: (0, utils_1.getBase58AddressFromHex)(owner_address) });
        }
        if (votes) {
            this.setVotes(votes.map((vote) => ({
                ...vote,
                vote_address: (0, utils_1.getBase58AddressFromHex)(vote.vote_address),
            })));
        }
    }
    async buildImplementation() {
        this.createVoteWitnessContractTransaction();
        /** @inheritdoccreateTransaction */
        // This method must be extended on child classes
        if (this._signingKeys.length > 0) {
            this.applySignatures();
        }
        if (!this.transaction.id) {
            throw new sdk_core_1.BuildTransactionError('A valid transaction must have an id');
        }
        return Promise.resolve(this.transaction);
    }
    /**
     * Creates the vote witness transaction
     */
    createVoteWitnessContractTransaction() {
        const rawDataHex = this.getVoteWitnessRawDataHex();
        const rawData = (0, utils_1.decodeTransaction)(rawDataHex);
        const contract = rawData.contract[0];
        const contractParameter = contract.parameter;
        contractParameter.value.owner_address = this._ownerAddress.toLocaleLowerCase();
        contractParameter.value.votes = this._votes;
        contractParameter.type_url = 'type.googleapis.com/protocol.VoteWitnessContract';
        contract.type = 'VoteWitnessContract';
        const hexBuffer = Buffer.from(rawDataHex, 'hex');
        const id = (0, crypto_1.createHash)('sha256').update(hexBuffer).digest('hex');
        const txRecip = {
            raw_data: rawData,
            raw_data_hex: rawDataHex,
            txID: id,
            signature: this.transaction.signature,
        };
        this.transaction = new transaction_1.Transaction(this._coinConfig, txRecip);
    }
    /**
     * Helper method to get the vote witness transaction raw data hex
     * @returns vote witness transaction raw data hex
     */
    getVoteWitnessRawDataHex() {
        const voteArray = this._votes.map((vote) => {
            const voteObject = tron_1.protocol.Vote.fromObject({
                voteAddress: (0, utils_1.getByteArrayFromHexAddress)(vote.vote_address),
                voteCount: vote.vote_count,
            });
            return tron_1.protocol.Vote.create(voteObject);
        });
        const rawContract = {
            ownerAddress: (0, utils_1.getByteArrayFromHexAddress)(this._ownerAddress),
            votes: voteArray,
        };
        const voteWitnessContract = tron_1.protocol.VoteWitnessContract.fromObject(rawContract);
        const voteWitnessContractBytes = tron_1.protocol.VoteWitnessContract.encode(voteWitnessContract).finish();
        const txContract = {
            type: ContractType.VoteWitnessContract,
            parameter: {
                value: voteWitnessContractBytes,
                type_url: 'type.googleapis.com/protocol.VoteWitnessContract',
            },
        };
        const raw = {
            refBlockBytes: Buffer.from(this._refBlockBytes, 'hex'),
            refBlockHash: Buffer.from(this._refBlockHash, 'hex'),
            expiration: this._expiration || Date.now() + utils_1.TRANSACTION_DEFAULT_EXPIRATION,
            timestamp: this._timestamp || Date.now(),
            contract: [txContract],
        };
        const rawTx = tron_1.protocol.Transaction.raw.create(raw);
        return Buffer.from(tron_1.protocol.Transaction.raw.encode(rawTx).finish()).toString('hex');
    }
    /** @inheritdoc */
    signImplementation(key) {
        if (this._signingKeys.some((signingKey) => signingKey.key === key.key)) {
            throw new sdk_core_1.SigningError('Duplicated key');
        }
        this._signingKeys.push(key);
        // We keep this return for compatibility but is not meant to be use
        return this.transaction;
    }
    applySignatures() {
        if (!this.transaction.inputs) {
            throw new sdk_core_1.SigningError('Transaction has no inputs');
        }
        this._signingKeys.forEach((key) => this.applySignature(key));
    }
    validateTransaction(transaction) {
        this.validateVoteTransactionFields();
    }
    /**
     * Validates the votes array
     *
     * @param {VoteWitnessData[]} votes - The votes array to validate
     * @throws {Error} when the votes array is invalid
     */
    validateVotes(votes) {
        if (!votes || votes.length === 0) {
            throw new Error('Votes array is missing or empty');
        }
        votes.forEach((vote) => {
            if (!vote.vote_address) {
                throw new Error('Vote address is missing');
            }
            this.validateAddress({ address: vote.vote_address });
            if (vote.vote_count == null || vote.vote_count < 0) {
                throw new Error('Invalid vote count');
            }
        });
    }
    /**
     * Validates if the transaction is a valid vote transaction
     * @param {TransactionReceipt} transaction - The transaction to validate
     * @throws {InvalidTransactionError} when the transaction is invalid
     */
    validateVoteTransactionFields() {
        if (!this._ownerAddress) {
            throw new sdk_core_1.BuildTransactionError('Missing parameter: source');
        }
        if (!this._refBlockBytes || !this._refBlockHash) {
            throw new sdk_core_1.BuildTransactionError('Missing block reference information');
        }
        // Validate votes
        if (!this._votes || this._votes.length === 0) {
            throw new sdk_core_1.BuildTransactionError('Missing or empty votes array');
        }
    }
}
exports.VoteWitnessTxBuilder = VoteWitnessTxBuilder;
