// @bitgo/sdk-coin-trx/dist/src/lib/withdrawBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawBalanceTxBuilder = void 0;
const crypto_1 = require("crypto");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
const tron_1 = require("../../resources/protobuf/tron");
const utils_1 = require("./utils");
var ContractType = tron_1.protocol.Transaction.Contract.ContractType;
class WithdrawBalanceTxBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._signingKeys = [];
        this.transaction = new transaction_1.Transaction(_coinConfig);
    }
    /** @inheritdoc */
    get transactionType() {
        return sdk_core_1.TransactionType.StakingClaim;
    }
    /** @inheritdoc */
    extendValidTo(extensionMs) {
        if (this.transaction.signature && this.transaction.signature.length > 0) {
            throw new sdk_core_1.ExtendTransactionError('Cannot extend a signed transaction');
        }
        if (extensionMs <= 0) {
            throw new Error('Value cannot be below zero');
        }
        if (extensionMs > utils_1.TRANSACTION_MAX_EXPIRATION) {
            throw new sdk_core_1.ExtendTransactionError('The expiration cannot be extended more than one year');
        }
        if (this._expiration) {
            this._expiration = this._expiration + extensionMs;
        }
        else {
            throw new Error('There is not expiration to extend');
        }
    }
    initBuilder(rawTransaction) {
        this.transaction = this.fromImplementation(rawTransaction);
        this.transaction.setTransactionType(this.transactionType);
        this.validateRawTransaction(rawTransaction);
        const tx = this.fromImplementation(rawTransaction);
        this.transaction = tx;
        this._signingKeys = [];
        const rawData = tx.toJson().raw_data;
        this._refBlockBytes = rawData.ref_block_bytes;
        this._refBlockHash = rawData.ref_block_hash;
        this._expiration = rawData.expiration;
        this._timestamp = rawData.timestamp;
        const contractCall = rawData.contract[0];
        this.initWithdrawBalanceContractCall(contractCall);
        return this;
    }
    /**
     * Initialize the withdraw balance contract call specific data
     *
     * @param {WithdrawBalanceContract} withdrawBalanceContractCall object with freeze txn data
     */
    initWithdrawBalanceContractCall(withdrawBalanceContractCall) {
        const { owner_address } = withdrawBalanceContractCall.parameter.value;
        if (owner_address) {
            this.source({ address: (0, utils_1.getBase58AddressFromHex)(owner_address) });
        }
    }
    async buildImplementation() {
        this.createWithdrawBalanceTransaction();
        /** @inheritdoccreateTransaction */
        // This method must be extended on child classes
        if (this._signingKeys.length > 0) {
            this.applySignatures();
        }
        if (!this.transaction.id) {
            throw new sdk_core_1.BuildTransactionError('A valid transaction must have an id');
        }
        return Promise.resolve(this.transaction);
    }
    /**
     * Helper method to create the withdraw balance transaction
     */
    createWithdrawBalanceTransaction() {
        const rawDataHex = this.getWithdrawBalanceRawDataHex();
        const rawData = (0, utils_1.decodeTransaction)(rawDataHex);
        const contract = rawData.contract[0];
        const contractParameter = contract.parameter;
        contractParameter.value.owner_address = this._ownerAddress.toLocaleLowerCase();
        contractParameter.type_url = 'type.googleapis.com/protocol.WithdrawBalanceContract';
        contract.type = 'WithdrawBalanceContract';
        const hexBuffer = Buffer.from(rawDataHex, 'hex');
        const id = (0, crypto_1.createHash)('sha256').update(hexBuffer).digest('hex');
        const txRecip = {
            raw_data: rawData,
            raw_data_hex: rawDataHex,
            txID: id,
            signature: this.transaction.signature,
        };
        this.transaction = new transaction_1.Transaction(this._coinConfig, txRecip);
    }
    /**
     * Helper method to get the withdraw expire unfreeze transaction raw data hex
     *
     * @returns {string} the freeze balance transaction raw data hex
     */
    getWithdrawBalanceRawDataHex() {
        const rawContract = {
            ownerAddress: (0, utils_1.getByteArrayFromHexAddress)(this._ownerAddress),
        };
        const withdrawBalanceContract = tron_1.protocol.WithdrawBalanceContract.fromObject(rawContract);
        const withdrawBalanceContractBytes = tron_1.protocol.WithdrawBalanceContract.encode(withdrawBalanceContract).finish();
        const txContract = {
            type: ContractType.WithdrawBalanceContract,
            parameter: {
                value: withdrawBalanceContractBytes,
                type_url: 'type.googleapis.com/protocol.WithdrawBalanceContract',
            },
        };
        const raw = {
            refBlockBytes: Buffer.from(this._refBlockBytes, 'hex'),
            refBlockHash: Buffer.from(this._refBlockHash, 'hex'),
            expiration: this._expiration || Date.now() + utils_1.TRANSACTION_DEFAULT_EXPIRATION,
            timestamp: this._timestamp || Date.now(),
            contract: [txContract],
        };
        const rawTx = tron_1.protocol.Transaction.raw.create(raw);
        return Buffer.from(tron_1.protocol.Transaction.raw.encode(rawTx).finish()).toString('hex');
    }
    /** @inheritdoc */
    signImplementation(key) {
        if (this._signingKeys.some((signingKey) => signingKey.key === key.key)) {
            throw new sdk_core_1.SigningError('Duplicated key');
        }
        this._signingKeys.push(key);
        // We keep this return for compatibility but is not meant to be use
        return this.transaction;
    }
    applySignatures() {
        if (!this.transaction.inputs) {
            throw new sdk_core_1.SigningError('Transaction has no inputs');
        }
        this._signingKeys.forEach((key) => this.applySignature(key));
    }
    /**
     * Validates the transaction
     *
     * @param {Transaction} transaction - The transaction to validate
     * @throws {BuildTransactionError} when the transaction is invalid
     */
    validateTransaction(transaction) {
        this.validateWithdrawBalanceTransactionFields();
    }
    /**
     * Validates if the transaction is a valid withdraw balance transaction
     *
     * @param {TransactionReceipt} transaction - The transaction to validate
     * @throws {BuildTransactionError} when the transaction is invalid
     */
    validateWithdrawBalanceTransactionFields() {
        if (!this._ownerAddress) {
            throw new sdk_core_1.BuildTransactionError('Missing parameter: source');
        }
        if (!this._refBlockBytes || !this._refBlockHash) {
            throw new sdk_core_1.BuildTransactionError('Missing block reference information');
        }
    }
}
exports.WithdrawBalanceTxBuilder = WithdrawBalanceTxBuilder;
