// @bitgo/sdk-coin-trx/dist/src/lib/wrappedBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrappedBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const keyPair_1 = require("./keyPair");
const utils_1 = require("./utils");
const enum_1 = require("./enum");
const contractCallBuilder_1 = require("./contractCallBuilder");
const tokenTransferBuilder_1 = require("./tokenTransferBuilder");
const freezeBalanceTxBuilder_1 = require("./freezeBalanceTxBuilder");
const voteWitnessTxBuilder_1 = require("./voteWitnessTxBuilder");
const unfreezeBalanceTxBuilder_1 = require("./unfreezeBalanceTxBuilder");
const withdrawExpireUnfreezeTxBuilder_1 = require("./withdrawExpireUnfreezeTxBuilder");
const withdrawBuilder_1 = require("./withdrawBuilder");
const delegateResourceTxBuilder_1 = require("./delegateResourceTxBuilder");
const undelegateResourceTxBuilder_1 = require("./undelegateResourceTxBuilder");
/**
 * Wrapped Builder class
 * This builder is created to maintain compatibility with the current uses of account-lib
 * It has an instance of Transaction Builder or Contract Call Builder as required.
 */
class WrappedBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        // defaults to old builder
        this._builder = new transactionBuilder_1.TransactionBuilder(_coinConfig);
    }
    /**
     * Returns a specific builder to create a contract call transaction
     *
     * @param {Transaction} [tx] The transaction to initialize builder
     * @returns {ContractCallBuilder} The specific contract call builder
     */
    getContractCallBuilder(tx) {
        return this.initializeBuilder(tx, new contractCallBuilder_1.ContractCallBuilder(this._coinConfig));
    }
    getTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new transactionBuilder_1.TransactionBuilder(this._coinConfig));
    }
    getTokenTransferBuilder(tx) {
        return this.initializeBuilder(tx, new tokenTransferBuilder_1.TokenTransferBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create a freeze balance transaction
     *
     * @param {TransactionReceipt} transaction - The transaction to initialize builder
     * @returns {FreezeBalanceTxBuilder} The specific freeze balance builder
     */
    getFreezeBalanceV2TxBuilder(tx) {
        return this.initializeBuilder(tx, new freezeBalanceTxBuilder_1.FreezeBalanceTxBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create a vote witness transaction
     *
     * @param {TransactionReceipt} transaction - The transaction to initialize builder
     * @returns {VoteWitnessTxBuilder} The specific vote witness builder
     */
    getVoteWitnessTxBuilder(tx) {
        return this.initializeBuilder(tx, new voteWitnessTxBuilder_1.VoteWitnessTxBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create an unfreeze balance transaction
     *
     * @param {Transaction} [tx] The transaction to initialize builder
     * @returns {UnfreezeBalanceTxBuilder} The specific unfreeze builder
     */
    getUnfreezeBalanceV2TxBuilder(tx) {
        return this.initializeBuilder(tx, new unfreezeBalanceTxBuilder_1.UnfreezeBalanceTxBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create a withdraw expire unfreeze transaction
     *
     * @param {Transaction} [tx] The transaction to initialize builder
     * @returns {WithdrawExpireUnfreezeTxBuilder} The specific withdraw builder
     */
    getWithdrawExpireUnfreezeTxBuilder(tx) {
        return this.initializeBuilder(tx, new withdrawExpireUnfreezeTxBuilder_1.WithdrawExpireUnfreezeTxBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create a withdraw balance transaction
     *
     * @param {Transaction} [tx] The transaction to initialize builder
     * @returns {WithdrawBalanceTxBuilder} The specific withdraw balance builder
     */
    getWithdrawBalanceTxBuilder(tx) {
        return this.initializeBuilder(tx, new withdrawBuilder_1.WithdrawBalanceTxBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create a delegate resource transaction
     *
     * @param {TransactionReceipt | string} [tx] The transaction to initialize builder
     * @returns {DelegateResourceTxBuilder} The specific delegate resource builder
     */
    getDelegateResourceTxBuilder(tx) {
        return this.initializeBuilder(tx, new delegateResourceTxBuilder_1.DelegateResourceTxBuilder(this._coinConfig));
    }
    /**
     * Returns a specific builder to create a undelegate resource transaction
     *
     * @param {TransactionReceipt | string} [tx] The transaction to initialize builder
     * @returns {UndelegateResourceTxBuilder} The specific delegate resource builder
     */
    getUnDelegateResourceTxBuilder(tx) {
        return this.initializeBuilder(tx, new undelegateResourceTxBuilder_1.UndelegateResourceTxBuilder(this._coinConfig));
    }
    initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
    /** @inheritdoc */
    extendValidTo(extensionMs) {
        this._builder.extendValidTo(extensionMs);
    }
    /** @inheritdoc */
    sign(key) {
        this._builder.sign(key);
    }
    /** @inheritDoc */
    addSignature(publicKey, signature) {
        this._builder.addSignature(publicKey, signature);
    }
    /** @inheritdoc */
    async build() {
        return this._builder.build();
    }
    /** @inheritdoc */
    from(raw) {
        this.validateRawTransaction(raw);
        const rawDataHex = this.getTxReceipt(raw);
        const decodedTx = (0, utils_1.decodeTransaction)(rawDataHex);
        const contractType = decodedTx.contractType;
        switch (contractType) {
            case enum_1.ContractType.Transfer:
            case enum_1.ContractType.AccountPermissionUpdate:
                this._builder = this.getTransactionBuilder(raw);
                return this._builder;
            case enum_1.ContractType.TriggerSmartContract:
                return this.getContractCallBuilder(raw);
            case enum_1.ContractType.FreezeBalanceV2:
                return this.getFreezeBalanceV2TxBuilder(raw);
            case enum_1.ContractType.VoteWitness:
                return this.getVoteWitnessTxBuilder(raw);
            case enum_1.ContractType.UnfreezeBalanceV2:
                return this.getUnfreezeBalanceV2TxBuilder(raw);
            case enum_1.ContractType.WithdrawExpireUnfreeze:
                return this.getWithdrawExpireUnfreezeTxBuilder(raw);
            case enum_1.ContractType.WithdrawBalance:
                return this.getWithdrawBalanceTxBuilder(raw);
            case enum_1.ContractType.DelegateResourceContract:
                return this.getDelegateResourceTxBuilder(raw);
            case enum_1.ContractType.UnDelegateResourceContract:
                return this.getUnDelegateResourceTxBuilder(raw);
            default:
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction type: ' + contractType);
        }
    }
    /**
     * Get the raw data hex from a raw transaction
     *
     * @param {string | { [key: string]: any }} raw the raw transaction as a string or as an object
     * @returns {string} the raw data hex
     */
    getTxReceipt(raw) {
        return raw['raw_data_hex'] || this.getTxReceipt(JSON.parse(raw));
    }
    /** @inheritdoc */
    validateAddress(address) {
        this._builder.validateAddress(address);
    }
    /** @inheritdoc */
    validateKey(key) {
        try {
            new keyPair_1.KeyPair({ prv: key.key });
        }
        catch (err) {
            throw new Error('The provided key is not valid');
        }
    }
    /** @inheritdoc */
    validateRawTransaction(rawTransaction) {
        this._builder.validateRawTransaction(rawTransaction);
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        this._builder.validateTransaction(transaction);
    }
    /** @inheritdoc */
    validateValue(value) {
        this._builder.validateValue(value);
    }
}
exports.WrappedBuilder = WrappedBuilder;
