// @bitgo/sdk-core/dist/src/account-lib/baseCoin/baseTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseTransaction = void 0;
const errors_1 = require("./errors");
/**
 * Generic transaction to be extended with coin specific logic.
 */
class BaseTransaction {
    /**
     * Base constructor.
     *
     * @param _coinConfig BaseCoin from statics library
     */
    constructor(_coinConfig) {
        this._coinConfig = _coinConfig;
        this._inputs = [];
        this._outputs = [];
        this._signatures = [];
        this._id = undefined;
        this._type = undefined;
    }
    /**
     * Get the transaction id as seen in the blockchain. Transactions computed offline may not have an
     * id, however, this is left to the coin implementation.
     */
    get id() {
        return this._id;
    }
    /**
     * One of {@link TransactionType}
     */
    get type() {
        return this._type;
    }
    /**
     * Get the list of outputs. Amounts are expressed in absolute value.
     */
    get outputs() {
        return this._outputs;
    }
    /**
     * Get the list of inputs. Amounts are expressed in absolute value.
     */
    get inputs() {
        return this._inputs;
    }
    /**
     * Get the list of signatures (if any) produced for this transaction.
     */
    get signature() {
        return this._signatures;
    }
    /**
     * Returns the portion of the transaction that needs to be signed in Buffer format.
     * Only needed for coins that support adding signatures directly (e.g. TSS).
     */
    get signablePayload() {
        throw new errors_1.NotImplementedError('signablePayload not implemented');
    }
    /**
     * Explain/parse a given coin transaction
     * TODO: Move all previous explainTransactions from 'core' to 'account-lib' for other coins,
     * TODO: convert to abstract
     */
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    explainTransaction() { }
}
exports.BaseTransaction = BaseTransaction;
