// @bitgo/sdk-core/dist/src/account-lib/baseCoin/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeSignature = serializeSignature;
exports.serializeSignatures = serializeSignatures;
exports.deserializeSignature = deserializeSignature;
exports.deserializeSignatures = deserializeSignatures;
exports.isSeed = isSeed;
exports.isPrivateKey = isPrivateKey;
exports.isPublicKey = isPublicKey;
function serializeSignature(signature) {
    return {
        publicKey: signature.publicKey.pub,
        signature: signature.signature.toString('base64'),
    };
}
function serializeSignatures(signatures) {
    return signatures.map(serializeSignature);
}
function deserializeSignature(serialized) {
    return {
        publicKey: { pub: serialized.publicKey },
        signature: Buffer.from(serialized.signature, 'base64'),
    };
}
function deserializeSignatures(serialized) {
    return serialized?.map(deserializeSignature) || [];
}
/**
 * @param source
 */
function isSeed(source) {
    return source.seed !== undefined;
}
/**
 * @param source
 */
function isPrivateKey(source) {
    return source.prv !== undefined;
}
/**
 * @param source
 */
function isPublicKey(source) {
    return source.pub !== undefined;
}
