// @bitgo/sdk-core/dist/src/bitgo/tss/eddsa/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.encryptYShare = exports.sendSignatureShare = exports.getTxRequest = exports.sendUserToBitgoGShare = exports.getBitgoToUserRShare = exports.offerUserToBitgoRShare = exports.createUserToBitGoGShare = exports.createUserSignShare = exports.createCombinedKey = exports.EDDSAMethodTypes = void 0;
const EDDSAMethods = __importStar(require("./eddsa"));
exports.EDDSAMethodTypes = __importStar(require("./types"));
exports.default = EDDSAMethods;
// exporting this types for backward compatibility.
var eddsa_1 = require("./eddsa");
Object.defineProperty(exports, "createCombinedKey", { enumerable: true, get: function () { return eddsa_1.createCombinedKey; } });
Object.defineProperty(exports, "createUserSignShare", { enumerable: true, get: function () { return eddsa_1.createUserSignShare; } });
Object.defineProperty(exports, "createUserToBitGoGShare", { enumerable: true, get: function () { return eddsa_1.createUserToBitGoGShare; } });
Object.defineProperty(exports, "offerUserToBitgoRShare", { enumerable: true, get: function () { return eddsa_1.offerUserToBitgoRShare; } });
Object.defineProperty(exports, "getBitgoToUserRShare", { enumerable: true, get: function () { return eddsa_1.getBitgoToUserRShare; } });
Object.defineProperty(exports, "sendUserToBitgoGShare", { enumerable: true, get: function () { return eddsa_1.sendUserToBitgoGShare; } });
Object.defineProperty(exports, "getTxRequest", { enumerable: true, get: function () { return eddsa_1.getTxRequest; } });
Object.defineProperty(exports, "sendSignatureShare", { enumerable: true, get: function () { return eddsa_1.sendSignatureShare; } });
Object.defineProperty(exports, "encryptYShare", { enumerable: true, get: function () { return eddsa_1.encryptYShare; } });
