// bitgo/dist/test/v2/unit/internal/tssUtils/ecdsaMPCv2/signTxRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBitGoPartyGpgKeyPrv = getBitGoPartyGpgKeyPrv;
exports.getUserPartyGpgKeyPublic = getUserPartyGpgKeyPublic;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const fs = require("fs");
const common_1 = require("../common");
const openpgp = require("openpgp");
const nock = require("nock");
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../../../../../src");
const createKeccakHash = require('keccak');
describe('signTxRequest:', function () {
    let tssUtils;
    let wallet;
    let bitgo;
    let baseCoin;
    let bitgoGpgKey;
    const coinName = 'hteth';
    const reqId = new sdk_core_1.RequestTracer();
    const txRequestId = 'randomTxReqId';
    const signableHex = 'e27aecaea559fbedc9ae8a22b0ab6654c2d686403c2aeb434b302545c94eed3b';
    const txRequest = {
        txRequestId,
        enterpriseId: '4517abfb-f567-4b7a-9f91-407509d29403',
        transactions: [
            {
                unsignedTx: {
                    serializedTxHex: 'TOO MANY SECRETS',
                    signableHex,
                    derivationPath: 'm/0', // Needs this when key derivation is supported
                },
                state: 'pendingSignature',
                signatureShares: [],
            },
        ],
        unsignedTxs: [],
        date: new Date().toISOString(),
        intent: {
            intentType: 'payment',
        },
        latest: true,
        state: 'pendingUserSignature',
        walletType: 'hot',
        walletId: 'walletId',
        policiesChecked: true,
        version: 1,
        userId: 'userId',
        apiVersion: 'full',
    };
    const txRequestForMessageSigning = {
        txRequestId,
        enterpriseId: '4517abfb-f567-4b7a-9f91-407509d29403',
        messages: [
            {
                messageRaw: 'TOO MANY SECRETS',
                derivationPath: 'm/0',
                state: 'pendingSignature',
                signatureShares: [],
            },
        ],
        unsignedTxs: [],
        date: new Date().toISOString(),
        intent: {
            intentType: 'payment',
        },
        latest: true,
        state: 'pendingUserSignature',
        walletType: 'hot',
        walletId: 'walletId',
        policiesChecked: true,
        version: 1,
        userId: 'userId',
        apiVersion: 'full',
    };
    const vector = {
        party1: 0,
        party2: 2,
        party3: 1,
    };
    // To generate the fixtures, run DKG as in the dklsDkg.ts tests and save the resulting party.getKeyShare in a file by doing fs.writeSync(party.getKeyShare()).
    const shareFiles = [
        `${__dirname}/fixtures/userShare`,
        `${__dirname}/fixtures/backupShare`,
        `${__dirname}/fixtures/bitgoShare`,
    ];
    let bitgoParty;
    before(async () => {
        bitgo = sdk_test_1.TestBitGo.decorate(src_1.BitGo, { env: 'mock' });
        bitgo.initializeTestVars();
        const bgUrl = sdk_core_1.common.Environments[bitgo.getEnv()].uri;
        bitgoGpgKey = await openpgp.generateKey({
            userIDs: [
                {
                    name: 'bitgo',
                },
            ],
            curve: 'secp256k1',
            config: {
                rejectCurves: new Set(),
            },
        });
        const constants = {
            mpc: {
                bitgoPublicKey: bitgoGpgKey.publicKey,
                bitgoMPCv2PublicKey: bitgoGpgKey.publicKey,
            },
        };
        nock(bgUrl).get('/api/v1/client/constants').times(20).reply(200, { ttl: 3600, constants });
        baseCoin = bitgo.coin(coinName);
        let hashFn;
        try {
            hashFn = baseCoin.getHashFunction();
        }
        catch (err) {
            hashFn = createKeccakHash('keccak256');
        }
        const hashBuffer = hashFn.update(Buffer.from(signableHex, 'hex')).digest();
        // Nock out both the user and bitgo side responses to create valid signatures
        bitgoParty = new sdk_lib_mpc_1.DklsDsg.Dsg(fs.readFileSync(shareFiles[vector.party2]), vector.party2, txRequest.transactions[0].unsignedTx.derivationPath, hashBuffer);
        // // Round 1 ////
        const walletData = {
            id: txRequest.walletId,
            enterprise: txRequest.enterpriseId,
            coin: coinName,
            coinSpecific: {},
            multisigType: 'tss',
            multisigTypeVersion: 'MPCv2',
        };
        wallet = new sdk_core_1.Wallet(bitgo, baseCoin, walletData);
        tssUtils = new sdk_core_1.ECDSAUtils.EcdsaMPCv2Utils(bitgo, baseCoin, wallet);
    });
    beforeEach(async function () {
        await nockGetBitgoPublicKeyBasedOnFeatureFlags(coinName, txRequest.enterpriseId, bitgoGpgKey);
    });
    after(function () {
        nock.cleanAll();
    });
    afterEach(function () {
        bitgoParty.endSession();
        nock.cleanAll();
    });
    it('successfully signs a txRequest with user key for a dkls hot wallet with WP', async function () {
        const nockPromises = [
            await nockTxRequestResponseSignatureShareRoundOne(bitgoParty, txRequest, bitgoGpgKey),
            await nockTxRequestResponseSignatureShareRoundTwo(bitgoParty, txRequest, bitgoGpgKey),
            await nockTxRequestResponseSignatureShareRoundThree(txRequest),
            await nockSendTxRequest(txRequest),
        ];
        await Promise.all(nockPromises);
        const userShare = fs.readFileSync(shareFiles[vector.party1]);
        const userPrvBase64 = Buffer.from(userShare).toString('base64');
        await tssUtils.signTxRequest({
            txRequest,
            prv: userPrvBase64,
            reqId,
        });
        nockPromises[0].isDone().should.be.true();
        nockPromises[1].isDone().should.be.true();
        nockPromises[2].isDone().should.be.true();
    });
    it('successfully signs a txRequest with backup key for a dkls hot wallet with WP', async function () {
        const nockPromises = [
            await nockTxRequestResponseSignatureShareRoundOne(bitgoParty, txRequest, bitgoGpgKey, 1),
            await nockTxRequestResponseSignatureShareRoundTwo(bitgoParty, txRequest, bitgoGpgKey, 1),
            await nockTxRequestResponseSignatureShareRoundThree(txRequest),
            await nockSendTxRequest(txRequest),
        ];
        await Promise.all(nockPromises);
        const backupShare = fs.readFileSync(shareFiles[vector.party3]);
        const backupPrvBase64 = Buffer.from(backupShare).toString('base64');
        await tssUtils.signTxRequest({
            txRequest,
            prv: backupPrvBase64,
            mpcv2PartyId: 1,
            reqId,
        });
        nockPromises[0].isDone().should.be.true();
        nockPromises[1].isDone().should.be.true();
        nockPromises[2].isDone().should.be.true();
    });
    it('successfully signs a txRequest with a message for a dkls hot wallet with WP', async function () {
        const nockPromises = [
            await nockTxRequestResponseSignatureShareRoundOne(bitgoParty, txRequestForMessageSigning, bitgoGpgKey),
            await nockTxRequestResponseSignatureShareRoundTwo(bitgoParty, txRequestForMessageSigning, bitgoGpgKey),
            await nockTxRequestResponseSignatureShareRoundThree(txRequestForMessageSigning),
            await nockSendTxRequest(txRequestForMessageSigning),
        ];
        await Promise.all(nockPromises);
        const userShare = fs.readFileSync(shareFiles[vector.party1]);
        const userPrvBase64 = Buffer.from(userShare).toString('base64');
        await tssUtils.signTxRequest({
            txRequest,
            prv: userPrvBase64,
            reqId,
        });
        nockPromises[0].isDone().should.be.true();
        nockPromises[1].isDone().should.be.true();
        nockPromises[2].isDone().should.be.true();
    });
    it('successfully signs a txRequest for a dkls hot wallet after receiving multiple 429 errors', async function () {
        const nockPromises = [
            await nockTxRequestResponseSignatureShareRoundOne(bitgoParty, txRequest, bitgoGpgKey),
            await nockTxRequestResponseSignatureShareRoundTwo(bitgoParty, txRequest, bitgoGpgKey, 0, 3),
            await nockTxRequestResponseSignatureShareRoundThree(txRequest),
            await nockSendTxRequest(txRequest),
        ];
        await Promise.all(nockPromises);
        const userShare = fs.readFileSync(shareFiles[vector.party1]);
        const userPrvBase64 = Buffer.from(userShare).toString('base64');
        await tssUtils.signTxRequest({
            txRequest,
            prv: userPrvBase64,
            reqId,
        });
        nockPromises[0].isDone().should.be.true();
        nockPromises[1].isDone().should.be.true();
        nockPromises[2].isDone().should.be.true();
    });
    it('fails to signs a txRequest for a dkls hot wallet after receiving over 3 429 errors', async function () {
        const nockPromises = [
            await nockTxRequestResponseSignatureShareRoundOne(bitgoParty, txRequest, bitgoGpgKey),
            await nockTxRequestResponseSignatureShareRoundTwo(bitgoParty, txRequest, bitgoGpgKey, 0, 4),
        ];
        await Promise.all(nockPromises);
        const userShare = fs.readFileSync(shareFiles[vector.party1]);
        const userPrvBase64 = Buffer.from(userShare).toString('base64');
        await tssUtils
            .signTxRequest({
            txRequest,
            prv: userPrvBase64,
            reqId,
        })
            .should.be.rejectedWith('Too many requests, slow down!');
        nockPromises[0].isDone().should.be.true();
        nockPromises[1].isDone().should.be.false();
    });
});
function getBitGoPartyGpgKeyPrv(key) {
    return {
        partyId: 2,
        gpgKey: key.privateKey,
    };
}
function getUserPartyGpgKeyPublic(userPubKey, partyId = 0) {
    return {
        partyId: partyId,
        gpgKey: userPubKey,
    };
}
async function nockTxRequestResponseSignatureShareRoundOne(bitgoSession, txRequest, bitgoGpgKey, partyId = 0) {
    const transactions = (0, common_1.getRoute)('ecdsa');
    return nock('https://bitgo.fakeurl')
        .persist(true)
        .post(`/api/v2/wallet/${txRequest.walletId}/txrequests/${txRequest.txRequestId + transactions}/sign`, (body) => JSON.parse(body.signatureShares[0].share).type === 'round1Input')
        .times(1)
        .reply(200, async (uri, body) => {
        // Do the actual signing on BitGo's side based on User's messages
        const signatureShare = JSON.parse(body.signatureShares[0].share);
        const deserializedMessages = sdk_lib_mpc_1.DklsTypes.deserializeMessages({
            p2pMessages: [],
            broadcastMessages: [
                {
                    from: signatureShare.data.msg1.from,
                    payload: signatureShare.data.msg1.message,
                },
            ],
        });
        if (signatureShare.type === 'round1Input') {
            const bitgoToUserRound1BroadcastMsg = await bitgoSession.init();
            const bitgoToUserRound2Msg = bitgoSession.handleIncomingMessages({
                p2pMessages: [],
                broadcastMessages: deserializedMessages.broadcastMessages,
            });
            const serializedBitGoToUserRound1And2Msgs = sdk_lib_mpc_1.DklsTypes.serializeMessages({
                p2pMessages: bitgoToUserRound2Msg.p2pMessages,
                broadcastMessages: [bitgoToUserRound1BroadcastMsg],
            });
            const authEncMessages = await sdk_lib_mpc_1.DklsComms.encryptAndAuthOutgoingMessages(serializedBitGoToUserRound1And2Msgs, [getUserPartyGpgKeyPublic(body.signerGpgPublicKey, partyId)], [getBitGoPartyGpgKeyPrv(bitgoGpgKey)]);
            const bitgoToUserSignatureShare = {
                type: 'round1Output',
                data: {
                    msg1: {
                        from: authEncMessages.broadcastMessages[0].from,
                        signature: authEncMessages.broadcastMessages[0].payload.signature,
                        message: authEncMessages.broadcastMessages[0].payload.message,
                    },
                    msg2: {
                        from: authEncMessages.p2pMessages[0].from,
                        to: authEncMessages.p2pMessages[0].to,
                        encryptedMessage: authEncMessages.p2pMessages[0].payload.encryptedMessage,
                        signature: authEncMessages.p2pMessages[0].payload.signature,
                    },
                },
            };
            return {
                txRequestId: txRequest.txRequestId,
                transactions: [
                    {
                        signatureShares: [
                            {
                                from: sdk_core_1.SignatureShareType.BITGO,
                                to: partyId === 0 ? sdk_core_1.SignatureShareType.USER : sdk_core_1.SignatureShareType.BACKUP,
                                share: JSON.stringify(bitgoToUserSignatureShare),
                            },
                        ],
                    },
                ],
            };
        }
    });
}
async function nockTxRequestResponseSignatureShareRoundTwo(bitgoSession, txRequest, bitgoGpgKey, partyId = 0, rateLimitErrorCount = 0) {
    const transactions = (0, common_1.getRoute)('ecdsa');
    const scope = nock('https://bitgo.fakeurl');
    if (rateLimitErrorCount > 0) {
        scope
            .post(`/api/v2/wallet/${txRequest.walletId}/txrequests/${txRequest.txRequestId + transactions}/sign`, (body) => JSON.parse(body.signatureShares[0].share).type === 'round2Input')
            .times(rateLimitErrorCount)
            .reply(429, {
            error: 'Too many requests, slow down!',
            name: 'TooManyRequests',
            requestId: 'cm5qx01lh0013b2ek2sxl4w00',
            context: {},
        });
    }
    return scope
        .post(`/api/v2/wallet/${txRequest.walletId}/txrequests/${txRequest.txRequestId + transactions}/sign`, (body) => JSON.parse(body.signatureShares[0].share).type === 'round2Input')
        .times(1)
        .reply(200, async (uri, body) => {
        // Do the actual signing on BitGo's side based on User's messages
        const parsedSignatureShare = JSON.parse(body.signatureShares[0].share);
        const serializedMessages = await sdk_lib_mpc_1.DklsComms.decryptAndVerifyIncomingMessages({
            p2pMessages: [
                {
                    from: parsedSignatureShare.data.msg2.from,
                    to: parsedSignatureShare.data.msg2.to,
                    payload: {
                        encryptedMessage: parsedSignatureShare.data.msg2.encryptedMessage,
                        signature: parsedSignatureShare.data.msg2.signature,
                    },
                },
                {
                    from: parsedSignatureShare.data.msg3.from,
                    to: parsedSignatureShare.data.msg3.to,
                    payload: {
                        encryptedMessage: parsedSignatureShare.data.msg3.encryptedMessage,
                        signature: parsedSignatureShare.data.msg3.signature,
                    },
                },
            ],
            broadcastMessages: [],
        }, [getUserPartyGpgKeyPublic(body.signerGpgPublicKey, partyId)], [getBitGoPartyGpgKeyPrv(bitgoGpgKey)]);
        const deserializedMessages = sdk_lib_mpc_1.DklsTypes.deserializeMessages({
            p2pMessages: [serializedMessages.p2pMessages[0]],
            broadcastMessages: [],
        });
        if (parsedSignatureShare.type === 'round2Input') {
            const bitgoToUserRound3Msg = bitgoSession.handleIncomingMessages(deserializedMessages);
            const serializedBitGoToUserRound3Msgs = sdk_lib_mpc_1.DklsTypes.serializeMessages(bitgoToUserRound3Msg);
            const authEncMessages = await sdk_lib_mpc_1.DklsComms.encryptAndAuthOutgoingMessages(serializedBitGoToUserRound3Msgs, [getUserPartyGpgKeyPublic(body.signerGpgPublicKey, partyId)], [getBitGoPartyGpgKeyPrv(bitgoGpgKey)]);
            const bitgoToUserSignatureShare = {
                type: 'round2Output',
                data: {
                    msg3: {
                        from: authEncMessages.p2pMessages[0].from,
                        to: authEncMessages.p2pMessages[0].to,
                        encryptedMessage: authEncMessages.p2pMessages[0].payload.encryptedMessage,
                        signature: authEncMessages.p2pMessages[0].payload.signature,
                    },
                },
            };
            return {
                txRequestId: txRequest.txRequestId,
                transactions: [
                    {
                        signatureShares: [
                            {
                                from: partyId === 0 ? sdk_core_1.SignatureShareType.USER : sdk_core_1.SignatureShareType.BACKUP,
                                to: sdk_core_1.SignatureShareType.BITGO,
                                share: 'some old share we dont care about',
                            },
                            {
                                from: sdk_core_1.SignatureShareType.BITGO,
                                to: partyId === 0 ? sdk_core_1.SignatureShareType.USER : sdk_core_1.SignatureShareType.BACKUP,
                                share: JSON.stringify(bitgoToUserSignatureShare),
                            },
                        ],
                    },
                ],
            };
        }
    });
}
async function nockTxRequestResponseSignatureShareRoundThree(txRequest) {
    const transactions = (0, common_1.getRoute)('ecdsa');
    return nock('https://bitgo.fakeurl')
        .post(`/api/v2/wallet/${txRequest.walletId}/txrequests/${txRequest.txRequestId + transactions}/sign`, (body) => JSON.parse(body.signatureShares[0].share).type === 'round3Input')
        .times(1)
        .reply(200, async (uri, body) => {
        // Do the actual signing on BitGo's side based on User's messages
        return {
            txRequestId: txRequest.txRequestId,
        };
    });
}
async function nockSendTxRequest(txRequest) {
    const transactions = (0, common_1.getRoute)('ecdsa');
    return nock('https://bitgo.fakeurl')
        .post(`/api/v2/wallet/${txRequest.walletId}/txrequests/${txRequest.txRequestId + transactions}/send`)
        .times(1)
        .reply(200, {
        txRequestId: txRequest.txRequestId,
    });
}
async function nockGetBitgoPublicKeyBasedOnFeatureFlags(coin, enterpriseId, bitgoGpgKeyPair) {
    const bitgoGPGPublicKeyResponse = {
        name: 'irrelevant',
        publicKey: bitgoGpgKeyPair.publicKey,
        mpcv2PublicKey: bitgoGpgKeyPair.publicKey,
        enterpriseId,
    };
    nock('https://bitgo.fakeurl')
        .get(`/api/v2/${coin}/tss/pubkey`)
        .times(4)
        .query({ enterpriseId })
        .reply(200, bitgoGPGPublicKeyResponse);
    return bitgoGPGPublicKeyResponse;
}
