// @bitgo/abstract-substrate/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionTypes = exports.MethodNames = exports.SectionNames = void 0;
/**
 * Section names for the transaction methods.
 */
var SectionNames;
(function (SectionNames) {
    SectionNames["Proxy"] = "proxy";
})(SectionNames || (exports.SectionNames = SectionNames = {}));
/**
 * Method names for the transaction method. Names change based on the type of transaction e.g 'bond' for the staking transaction
 *
 * This is implemented as a const object with string literals to allow for extension in derived modules.
 */
exports.MethodNames = {
    /**
     * Transfer the entire transferable balance from the caller account.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#transferalldest-multiaddress-keep_alive-bool
     */
    TransferAll: 'transferAll',
    /**
     * Same as the transfer call, but with a check that the transfer will not kill the origin account.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#transferkeepalivedest-multiaddress-value-compactu128
     */
    TransferKeepAlive: 'transferKeepAlive',
    /**
     * Transfer funds with an optional memo attached.
     * The memo allows adding context or metadata to the transaction, commonly used for recordkeeping or identification.
     *
     * @see https://developers.polymesh.network/sdk-docs/enums/Generated/Types/BalancesTx/#transferwithmemo
     */
    TransferWithMemo: 'transferWithMemo',
    AddStake: 'addStake',
    RemoveStake: 'removeStake',
    /**
     * Take the origin account as a stash and lock up value of its balance.
     */
    Bond: 'bond',
    /**
     * Add some extra amount that have appeared in the stash free_balance into the balance up for staking.
     */
    BondExtra: 'bondExtra',
    /**
     * Declare the desire to nominate targets for the origin controller.
     */
    Nominate: 'nominate',
    /**
     * Declare no desire to either validate or nominate.
     */
    Chill: 'chill',
    /**
     * Schedule a portion of the stash to be unlocked ready for transfer out after the bond period ends.
     */
    Unbond: 'unbond',
    /**
     * Remove any unlocked chunks from the unlocking queue from our management.
     */
    WithdrawUnbonded: 'withdrawUnbonded',
    /**
     * Send a batch of dispatch calls.
     */
    Batch: 'batch',
    /**
     * Send a batch of dispatch calls and atomically execute them.
     */
    BatchAll: 'batchAll',
    /**
     * Transfer stake from one validator to another.
     */
    TransferStake: 'transferStake',
    /**
     * Move stake from one hotkey to another.
     */
    MoveStake: 'moveStake',
};
var TransactionTypes;
(function (TransactionTypes) {
    TransactionTypes["TRANSFER"] = "transfer";
})(TransactionTypes || (exports.TransactionTypes = TransactionTypes = {}));
