// @bitgo/abstract-substrate/dist/src/lib/nativeTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeTransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const iface_1 = require("./iface");
const transactionBuilder_1 = require("./transactionBuilder");
const txnSchema_1 = require("./txnSchema");
const utils_1 = __importDefault(require("./utils"));
class NativeTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._sweepFreeBalance = false;
        this._keepAddressAlive = true;
    }
    /**
     *
     * Dispatch the given call from an account that the sender is authorised for through add_proxy.
     *
     * @returns {UnsignedTransaction} an unsigned Dot transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#proxy
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        let transferTx;
        if (this._sweepFreeBalance) {
            transferTx = txwrapper_polkadot_1.methods.balances.transferAll({
                dest: { id: this._to },
                keepAlive: this._keepAddressAlive,
            }, baseTxInfo.baseTxInfo, baseTxInfo.options);
        }
        else {
            transferTx = txwrapper_polkadot_1.methods.balances.transferKeepAlive({
                value: this._amount,
                dest: { id: this._to },
            }, baseTxInfo.baseTxInfo, baseTxInfo.options);
        }
        return transferTx;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    /**
     *
     * Set this to be a sweep transaction, using TransferAll with keepAlive set to true by default.
     * If keepAlive is false, the entire address will be swept (including the 1 DOT minimum).
     *
     * @param {boolean} keepAlive - keep the address alive after this sweep
     * @returns {TransferBuilder} This transfer builder.
     *
     * @see https://github.com/paritytech/txwrapper-core/blob/main/docs/modules/txwrapper_substrate_src.methods.balances.md#transferall
     */
    sweep(keepAlive) {
        this._sweepFreeBalance = true;
        if (keepAlive !== undefined) {
            this._keepAddressAlive = keepAlive;
        }
        return this;
    }
    /**
     *
     * The amount for transfer transaction.
     *
     * @param {string} amount
     * @returns {TransferBuilder} This transfer builder.
     *
     * @see https://wiki.polkadot.network/docs/build-protocol-info
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     *
     * The destination address for transfer transaction.
     *
     * @param {string} dest
     * @returns {TransferBuilder} This transfer builder.
     *
     * @see https://wiki.polkadot.network/docs/build-protocol-info
     */
    to({ address }) {
        this.validateAddress({ address });
        this._to = address;
        return this;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === iface_1.MethodNames.TransferKeepAlive) {
            const txMethod = decodedTxn.method.args;
            const amount = `${txMethod.value}`;
            const to = txMethod.dest.id;
            const validationResult = txnSchema_1.TransferTransactionSchema.validate({ amount, to });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Transfer Transaction validation failed: ${validationResult.error.message}`);
            }
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (!this._method || !this._method.args) {
            throw new sdk_core_1.InvalidTransactionError('Transaction method or args are undefined');
        }
        if (this._method.name === iface_1.MethodNames.TransferKeepAlive) {
            const txMethod = this._method.args;
            this.amount(txMethod.value);
            this.to({
                address: utils_1.default.decodeSubstrateAddress(txMethod.dest.id, this.getAddressFormat()),
            });
        }
        else if (this._method.name === iface_1.MethodNames.TransferAll) {
            this._sweepFreeBalance = true;
            const txMethod = this._method.args;
            this.sweep(txMethod.keepAlive);
            this.to({
                address: utils_1.default.decodeSubstrateAddress(txMethod.dest.id, this.getAddressFormat()),
            });
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method.name}. Expected a transferKeepAlive or a proxy transferKeepAlive transaction`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._to, this._amount);
    }
    validateFields(to, amount) {
        let validationResult;
        if (this._sweepFreeBalance) {
            validationResult = txnSchema_1.TransferAllTransactionSchema.validate({ to });
        }
        else {
            validationResult = txnSchema_1.TransferTransactionSchema.validate({ amount, to });
        }
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Transaction validation failed: ${validationResult.error.message}`);
        }
    }
}
exports.NativeTransferBuilder = NativeTransferBuilder;
