// @bitgo/abstract-substrate/dist/src/lib/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const keyring_1 = __importStar(require("@polkadot/keyring"));
const util_1 = require("@polkadot/util");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const keyPair_1 = require("./keyPair");
const utils_1 = __importDefault(require("./utils"));
const Extrinsic_1 = require("@polkadot/types/extrinsic/v4/Extrinsic");
const constants_1 = require("./constants");
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
    }
    /** @inheritdoc */
    canSign({ key }) {
        const kp = new keyPair_1.KeyPair({ prv: key });
        const addr = kp.getAddress(this.getAddressFormat());
        return addr === this._sender;
    }
    /**
     * Sign a substrate transaction and update the transaction hex
     *
     * @param {KeyPair} keyPair - ed signature
     */
    sign(keyPair) {
        if (!this._substrateTransaction) {
            throw new sdk_core_1.InvalidTransactionError('No transaction data to sign');
        }
        const { prv, pub } = keyPair.getKeys();
        if (!prv) {
            throw new sdk_core_1.SigningError('Missing private key');
        }
        const signingPayload = txwrapper_polkadot_1.construct.signingPayload(this._substrateTransaction, {
            registry: this._registry,
        });
        // Sign a payload. This operation should be performed on an offline device.
        const keyring = new keyring_1.default({ type: 'ed25519' });
        const secretKey = new Uint8Array(Buffer.from(prv, 'hex'));
        const publicKey = new Uint8Array(Buffer.from(pub, 'hex'));
        const signingKeyPair = keyring.addFromPair({ secretKey, publicKey });
        const txHex = utils_1.default.createSignedTx(signingKeyPair, signingPayload, this._substrateTransaction, {
            metadataRpc: this._substrateTransaction.metadataRpc,
            registry: this._registry,
        });
        // get signature from signed txHex generated above
        this._signatures = [utils_1.default.recoverSignatureFromRawTx(txHex, { registry: this._registry })];
        this._signedTransaction = txHex;
    }
    /**
     * Adds the signature to the Substrate Transaction
     * @param {string} signature
     */
    addSignature(signature) {
        this._signedTransaction = utils_1.default.serializeSignedTransaction(this._substrateTransaction, signature, this._substrateTransaction.metadataRpc, this._registry);
    }
    /**
     * Returns a serialized representation of this transaction with a fake signature attached which
     * can be used to estimate transaction fees.
     */
    fakeSign() {
        return utils_1.default.serializeSignedTransaction(this._substrateTransaction, Transaction.FAKE_SIGNATURE, this._substrateTransaction.metadataRpc, this._registry);
    }
    registry(registry) {
        this._registry = registry;
    }
    chainName(chainName) {
        this._chainName = chainName;
    }
    sender(sender) {
        this._sender = sender;
    }
    /** @inheritdoc */
    toBroadcastFormat() {
        if (!this._substrateTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        if (this._signedTransaction && this._signedTransaction.length > 0) {
            return this._signedTransaction;
        }
        else {
            return txwrapper_polkadot_1.construct.signingPayload(this._substrateTransaction, {
                registry: this._registry,
            });
        }
    }
    transactionSize() {
        return this.toBroadcastFormat().length / 2;
    }
    /** @inheritdoc */
    toJson() {
        if (!this._substrateTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        const decodedTx = (0, txwrapper_polkadot_1.decode)(this._substrateTransaction, {
            metadataRpc: this._substrateTransaction.metadataRpc,
            registry: this._registry,
            isImmortalEra: utils_1.default.isZeroHex(this._substrateTransaction.era),
        });
        const result = {
            id: txwrapper_polkadot_1.construct.txHash(this.toBroadcastFormat()),
            sender: decodedTx.address,
            referenceBlock: decodedTx.blockHash,
            blockNumber: decodedTx.blockNumber,
            genesisHash: decodedTx.genesisHash,
            nonce: decodedTx.nonce,
            specVersion: decodedTx.specVersion,
            transactionVersion: decodedTx.transactionVersion,
            eraPeriod: decodedTx.eraPeriod,
            chainName: this._chainName,
            tip: decodedTx.tip ? Number(decodedTx.tip) : 0,
        };
        const txMethod = decodedTx.method.args;
        if (this.type === sdk_core_1.TransactionType.Send) {
            if (utils_1.default.isTransfer(txMethod)) {
                const keypairDest = new keyPair_1.KeyPair({
                    pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.dest.id)).toString('hex'),
                });
                result.to = keypairDest.getAddress(this.getAddressFormat());
                result.amount = txMethod.value;
            }
            else if (utils_1.default.isTransferAll(txMethod)) {
                const keypairDest = new keyPair_1.KeyPair({
                    pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.dest.id)).toString('hex'),
                });
                result.to = keypairDest.getAddress(this.getAddressFormat());
                result.keepAlive = txMethod.keepAlive;
            }
            else {
                throw new sdk_core_1.ParseTransactionError(`Serializing unknown Transfer type parameters`);
            }
        }
        else if (this.type === sdk_core_1.TransactionType.StakingActivate) {
            if (utils_1.default.isAddStake(txMethod)) {
                const keypairDest = new keyPair_1.KeyPair({
                    pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.hotkey)).toString('hex'),
                });
                // hotkey address of validator
                result.to = keypairDest.getAddress(this.getAddressFormat());
                result.amount = txMethod.amountStaked.toString();
                result.netuid = txMethod.netuid;
            }
            else if (utils_1.default.isBond(txMethod)) {
                result.amount = txMethod.value;
                result.payee = typeof txMethod.payee === 'string' ? txMethod.payee : txMethod.payee.Account;
            }
            else if (utils_1.default.isBondExtra(txMethod)) {
                result.amount = txMethod.maxAdditional;
            }
        }
        else if (this.type === sdk_core_1.TransactionType.StakingDeactivate) {
            if (utils_1.default.isRemoveStake(txMethod)) {
                const keypairDest = new keyPair_1.KeyPair({
                    pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.hotkey)).toString('hex'),
                });
                // hotkey address of validator
                result.to = keypairDest.getAddress(this.getAddressFormat());
                result.amount = txMethod.amountUnstaked.toString();
                result.netuid = txMethod.netuid;
            }
            else if (utils_1.default.isUnbond(txMethod)) {
                result.amount = txMethod.value;
            }
            else if (utils_1.default.isWithdrawUnbonded(txMethod)) {
                result.numSlashingSpans = txMethod.numSlashingSpans;
            }
        }
        else if (this.type === sdk_core_1.TransactionType.Batch) {
            if (utils_1.default.isBatch(txMethod)) {
                result.batchCalls = txMethod.calls;
                // Extract amount from batch calls for display
                if (txMethod.calls && txMethod.calls.length === 2) {
                    const firstCall = txMethod.calls[0];
                    const secondCall = txMethod.calls[1];
                    if (firstCall.method === 'bond' && secondCall.method === 'nominate') {
                        // Staking batch: bond + nominate
                        const bondArgs = firstCall.args;
                        result.amount = bondArgs.value;
                    }
                    else if (firstCall.method === 'chill' && secondCall.method === 'unbond') {
                        // Unstaking batch: chill + unbond
                        const unbondArgs = secondCall.args;
                        result.amount = unbondArgs.value;
                    }
                }
            }
        }
        return result;
    }
    explainTransferTransaction(json, explanationResult) {
        return {
            ...explanationResult,
            outputs: [
                {
                    address: json.to?.toString() || '',
                    amount: json.amount?.toString() || '',
                },
            ],
        };
    }
    explainStakeTransaction(json, explanationResult) {
        return {
            ...explanationResult,
            outputs: [
                {
                    address: json.to?.toString() || '',
                    amount: json.amount?.toString() || '',
                },
            ],
        };
    }
    explainUnstakeTransaction(json, explanationResult) {
        return {
            ...explanationResult,
            outputs: [
                {
                    address: json.sender.toString() || '',
                    amount: json.amount?.toString() || '',
                },
            ],
        };
    }
    /** @inheritdoc */
    explainTransaction() {
        const result = this.toJson();
        const outputs = [];
        const explanationResult = {
            // txhash used to identify the transactions
            id: result.id,
            outputAmount: result.amount?.toString() || '0',
            changeAmount: '0',
            changeOutputs: [],
            outputs,
            fee: {
                fee: result.tip?.toString() || '',
                type: 'tip',
            },
            type: this.type,
        };
        switch (this.type) {
            case sdk_core_1.TransactionType.Send:
                return this.explainTransferTransaction(result, explanationResult);
            case sdk_core_1.TransactionType.StakingActivate:
                return this.explainStakeTransaction(result, explanationResult);
            case sdk_core_1.TransactionType.StakingDeactivate:
                return this.explainUnstakeTransaction(result, explanationResult);
            default:
                throw new sdk_core_1.InvalidTransactionError('Transaction type not supported');
        }
    }
    /**
     * Load the input and output data on this transaction.
     */
    loadInputsAndOutputs() {
        if (!this._substrateTransaction) {
            return;
        }
        const decodedTx = (0, txwrapper_polkadot_1.decode)(this._substrateTransaction, {
            metadataRpc: this._substrateTransaction.metadataRpc,
            registry: this._registry,
            isImmortalEra: utils_1.default.isZeroHex(this._substrateTransaction.era),
        });
        if (this.type === sdk_core_1.TransactionType.Send) {
            this.decodeInputsAndOutputsForSend(decodedTx);
        }
        else if (this.type === sdk_core_1.TransactionType.StakingActivate) {
            this.decodeInputsAndOutputsForStakingActivate(decodedTx);
        }
        else if (this.type === sdk_core_1.TransactionType.StakingDeactivate) {
            this.decodeInputsAndOutputsForStakingDeactivate(decodedTx);
        }
        else if (this.type === sdk_core_1.TransactionType.Batch) {
            this.decodeInputsAndOutputsForBatch(decodedTx);
        }
    }
    decodeInputsAndOutputsForSend(decodedTx) {
        const txMethod = decodedTx.method.args;
        let to;
        let value;
        let from;
        if (utils_1.default.isTransferAll(txMethod)) {
            const keypairDest = new keyPair_1.KeyPair({
                pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.dest.id)).toString('hex'),
            });
            to = keypairDest.getAddress(this.getAddressFormat());
            value = '0'; // substrate transferAll's do not deserialize amounts
            from = decodedTx.address;
        }
        else if (utils_1.default.isTransfer(txMethod)) {
            const keypairDest = new keyPair_1.KeyPair({
                pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.dest.id)).toString('hex'),
            });
            to = keypairDest.getAddress(this.getAddressFormat());
            value = txMethod.value;
            from = decodedTx.address;
        }
        else {
            throw new sdk_core_1.ParseTransactionError(`Loading inputs of unknown Transfer type parameters`);
        }
        this._outputs = [
            {
                address: to,
                value,
                coin: this._coinConfig.name,
            },
        ];
        this._inputs = [
            {
                address: from,
                value,
                coin: this._coinConfig.name,
            },
        ];
    }
    decodeInputsAndOutputsForStakingActivate(decodedTx) {
        const txMethod = decodedTx.method.args;
        let to;
        let value;
        let from;
        if (utils_1.default.isAddStake(txMethod)) {
            const keypairDest = new keyPair_1.KeyPair({
                pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.hotkey)).toString('hex'),
            });
            to = keypairDest.getAddress(this.getAddressFormat());
            value = txMethod.amountStaked.toString();
            from = decodedTx.address;
        }
        else if (utils_1.default.isBond(txMethod)) {
            to = decodedTx.address; // For bond, funds are locked in the same account
            value = txMethod.value;
            from = decodedTx.address;
        }
        else if (utils_1.default.isBondExtra(txMethod)) {
            to = decodedTx.address; // For bond extra, funds are locked in the same account
            value = txMethod.maxAdditional;
            from = decodedTx.address;
        }
        else {
            throw new sdk_core_1.ParseTransactionError(`Loading inputs of unknown StakingActivate type parameters`);
        }
        this._outputs = [
            {
                address: to,
                value,
                coin: this._coinConfig.name,
            },
        ];
        this._inputs = [
            {
                address: from,
                value,
                coin: this._coinConfig.name,
            },
        ];
    }
    decodeInputsAndOutputsForStakingDeactivate(decodedTx) {
        const txMethod = decodedTx.method.args;
        let to;
        let value;
        let from;
        if (utils_1.default.isRemoveStake(txMethod)) {
            const keypairDest = new keyPair_1.KeyPair({
                pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.hotkey)).toString('hex'),
            });
            to = keypairDest.getAddress(this.getAddressFormat());
            value = txMethod.amountUnstaked.toString();
            from = decodedTx.address;
        }
        else if (utils_1.default.isUnbond(txMethod)) {
            to = decodedTx.address; // For unbond, funds are unlocked from the same account
            value = txMethod.value;
            from = decodedTx.address;
        }
        else if (utils_1.default.isWithdrawUnbonded(txMethod)) {
            to = decodedTx.address; // For withdraw unbonded, funds are returned to the same account
            value = '0'; // Amount is not specified in withdraw unbonded
            from = decodedTx.address;
        }
        else {
            throw new sdk_core_1.ParseTransactionError(`Loading inputs of unknown StakingDeactivate type parameters`);
        }
        this._outputs = [
            {
                address: from,
                value,
                coin: this._coinConfig.name,
            },
        ];
        this._inputs = [
            {
                address: to,
                value,
                coin: this._coinConfig.name,
            },
        ];
    }
    decodeInputsAndOutputsForBatch(decodedTx) {
        const txMethod = decodedTx.method.args;
        const sender = decodedTx.address;
        this._inputs = [];
        this._outputs = [];
        if (utils_1.default.isBatch(txMethod)) {
            if (!txMethod.calls) {
                throw new sdk_core_1.InvalidTransactionError('failed to decode calls from batch transaction');
            }
            // Handle different types of batch operations
            let totalStakingValue = '0';
            let hasStakingOperations = false;
            let hasUnstakingOperations = false;
            for (const call of txMethod.calls) {
                // Handle both possible formats: simple method names or callIndex with registry lookup
                let methodName;
                if (typeof call.method === 'string') {
                    methodName = call.method;
                }
                else {
                    try {
                        const callIndex = call.method;
                        const decodedCall = this._registry.findMetaCall(new Uint8Array(Buffer.from(callIndex.replace('0x', ''), 'hex')));
                        methodName = decodedCall.method;
                    }
                    catch (e) {
                        methodName = call.method;
                    }
                }
                if (methodName === 'bond') {
                    const args = call.args;
                    const value = args.value || '0';
                    totalStakingValue = value;
                    hasStakingOperations = true;
                }
                else if (methodName === 'chill') {
                    hasUnstakingOperations = true;
                }
                else if (methodName === 'unbond') {
                    const args = call.args;
                    const value = args.value || '0';
                    totalStakingValue = value;
                    hasUnstakingOperations = true;
                }
            }
            // For staking batch operations (bond + nominate or bondExtra + nominate)
            if (hasStakingOperations && !hasUnstakingOperations) {
                this._inputs.push({
                    address: sender,
                    value: totalStakingValue,
                    coin: this._coinConfig.name,
                });
                this._outputs.push({
                    address: sender, // For staking, funds are locked in the same account
                    value: totalStakingValue,
                    coin: this._coinConfig.name,
                });
            }
            // For unstaking batch operations (chill + unbond)
            else if (hasUnstakingOperations && !hasStakingOperations) {
                this._inputs.push({
                    address: sender,
                    value: totalStakingValue,
                    coin: this._coinConfig.name,
                });
                this._outputs.push({
                    address: sender, // For unstaking, funds are unlocked from the same account
                    value: totalStakingValue,
                    coin: this._coinConfig.name,
                });
            }
        }
    }
    /**
     * Constructs a signed payload using construct.signTx
     * This method will be called during the build step if a TSS signature
     * is added and will set the signTransaction which is the txHex that will be broadcasted
     * As well as add the signature used to sign to the signature array in hex format
     *
     * @param {Buffer} signature The signature to be added to a substrate transaction
     */
    constructSignedPayload(signature) {
        // 0x00 means its an ED25519 signature
        const edSignature = `0x00${signature.toString('hex')}`;
        try {
            this._signedTransaction = txwrapper_polkadot_1.construct.signedTx(this._substrateTransaction, edSignature, {
                registry: this._registry,
                metadataRpc: this._substrateTransaction.metadataRpc,
            });
        }
        catch (e) {
            throw new sdk_core_1.SigningError(`Unable to sign transaction with signature ${edSignature} ` + e);
        }
        this._signatures = [signature.toString('hex')];
    }
    setTransaction(tx) {
        this._substrateTransaction = tx;
    }
    /** @inheritdoc **/
    get signablePayload() {
        const extrinsicPayload = this._registry.createType('ExtrinsicPayload', this._substrateTransaction, {
            version: Extrinsic_1.EXTRINSIC_VERSION,
        });
        return (0, util_1.u8aToBuffer)(extrinsicPayload.toU8a({ method: true }));
    }
    /**
     * Set the transaction type.
     *
     * @param {TransactionType} transactionType The transaction type to be set.
     */
    transactionType(transactionType) {
        this._type = transactionType;
    }
    getAddressFormat() {
        return constants_1.DEFAULT_SUBSTRATE_PREFIX;
    }
}
exports.Transaction = Transaction;
Transaction.FAKE_SIGNATURE = `0x${Buffer.from(new Uint8Array(256).fill(1)).toString('hex')}`;
