// @bitgo/abstract-substrate/dist/src/lib/txnSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoveStakeTransactionSchema = exports.TransferStakeTransactionSchema = exports.UnstakeTransactionSchema = exports.StakeTransactionSchema = exports.TransferWithMemoTransactionSchema = exports.TransferAllTransactionSchema = exports.TransferTransactionSchema = exports.SignedTransactionSchema = exports.SigningPayloadTransactionSchema = exports.BaseTransactionSchema = void 0;
const joi_1 = __importDefault(require("joi"));
const utils_1 = __importDefault(require("./utils"));
const addressSchema = joi_1.default.string().custom((addr) => utils_1.default.isValidAddress(addr));
exports.BaseTransactionSchema = joi_1.default.object({
    sender: addressSchema.required(),
    blockNumber: joi_1.default.number().required(),
    blockHash: joi_1.default.string().required(),
    genesisHash: joi_1.default.string().required(),
    specVersion: joi_1.default.number().required(),
    specName: joi_1.default.string().required(),
    transactionVersion: joi_1.default.number().required(),
    chainName: joi_1.default.string().required(),
    eraPeriod: joi_1.default.number().required(),
    nonce: joi_1.default.number().required(),
    tip: joi_1.default.number().optional(),
});
exports.SigningPayloadTransactionSchema = joi_1.default.object({
    eraPeriod: joi_1.default.number().optional(),
    blockHash: joi_1.default.string().required(),
    nonce: joi_1.default.number().required(),
    tip: joi_1.default.number().optional(),
});
exports.SignedTransactionSchema = joi_1.default.object({
    sender: addressSchema.required(),
    nonce: joi_1.default.number().required(),
    eraPeriod: joi_1.default.number().optional(),
    tip: joi_1.default.number().optional(),
});
exports.TransferTransactionSchema = joi_1.default.object({
    amount: joi_1.default.string().required(),
    to: addressSchema.required(),
});
exports.TransferAllTransactionSchema = joi_1.default.object({
    to: addressSchema.required(),
});
exports.TransferWithMemoTransactionSchema = joi_1.default.object({
    amount: joi_1.default.string().required(),
    to: addressSchema.required(),
    memo: joi_1.default.string().required(),
});
exports.StakeTransactionSchema = joi_1.default.object({
    amountStaked: joi_1.default.string().required(),
    hotkey: joi_1.default.string().required(),
    netuid: joi_1.default.string().required(),
});
exports.UnstakeTransactionSchema = joi_1.default.object({
    amountUnstaked: joi_1.default.string().required(),
    hotkey: joi_1.default.string().required(),
    netuid: joi_1.default.string().required(),
});
exports.TransferStakeTransactionSchema = joi_1.default.object({
    destinationColdkey: addressSchema.required(),
    hotkey: addressSchema.required(),
    originNetuid: joi_1.default.string().regex(/^\d+$/).required(),
    destinationNetuid: joi_1.default.string().regex(/^\d+$/).required(),
    alphaAmount: joi_1.default.string().required(),
});
exports.MoveStakeTransactionSchema = joi_1.default.object({
    originHotkey: addressSchema.required(),
    destinationHotkey: addressSchema.required(),
    originNetuid: joi_1.default.string().required(),
    destinationNetuid: joi_1.default.string().required(),
    alphaAmount: joi_1.default.string().required(),
});
