// @bitgo/abstract-substrate/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const keyring_1 = require("@polkadot/keyring");
const decode_1 = require("@polkadot/keyring/pair/decode");
const Extrinsic_1 = require("@polkadot/types/extrinsic/v4/Extrinsic");
const util_1 = require("@polkadot/util");
const util_crypto_1 = require("@polkadot/util-crypto");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const bs58_1 = __importDefault(require("bs58"));
const hi_base32_1 = __importDefault(require("hi-base32"));
const tweetnacl_1 = __importDefault(require("tweetnacl"));
const _1 = require(".");
const singletonRegistry_1 = require("./singletonRegistry");
class Utils {
    /** @inheritdoc */
    isValidAddress(address) {
        try {
            (0, keyring_1.encodeAddress)((0, util_1.isHex)(address) ? (0, util_1.hexToU8a)(address) : (0, keyring_1.decodeAddress)(address));
            return true;
        }
        catch (error) {
            return false;
        }
    }
    /** @inheritdoc */
    isValidBlockId(hash) {
        return (0, util_1.isHex)(hash, 256);
    }
    /** @inheritdoc */
    isValidPrivateKey(key) {
        try {
            const decodedPrv = (0, util_1.hexToU8a)(key);
            return decodedPrv.length === tweetnacl_1.default.sign.secretKeyLength / 2;
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    isValidPublicKey(key) {
        let pubKey = key;
        // convert base58 pub key to hex format
        // tss common pub is in base58 format and decodes to length of 32
        if ((0, sdk_core_1.isBase58)(pubKey, 32)) {
            const base58Decode = bs58_1.default.decode(pubKey);
            pubKey = base58Decode.toString('hex');
        }
        return (0, sdk_core_1.isValidEd25519PublicKey)(pubKey);
    }
    /** @inheritdoc */
    isValidSignature(signature) {
        const signatureU8a = (0, util_1.u8aToU8a)(signature);
        return [64, 65, 66].includes(signatureU8a.length);
    }
    /**
     * Verifies the signature on a given message
     *
     * @param {string} signedMessage the signed message for the signature
     * @param {string} signature the signature to verify
     * @param {string} address the address of the signer
     * @returns {boolean} whether the signature is valid or not
     */
    verifySignature(signedMessage, signature, address) {
        const publicKey = (0, keyring_1.decodeAddress)(address);
        const hexPublicKey = (0, util_1.u8aToHex)(publicKey);
        return (0, util_crypto_1.signatureVerify)(signedMessage, signature, hexPublicKey).isValid;
    }
    /** @inheritdoc */
    isValidTransactionId(txId) {
        return (0, util_1.isHex)(txId, 256);
    }
    /**
     * decodeSeed decodes a substrate seed
     *
     * @param {string} seed - the seed to be validated.
     * @returns {Seed} - the object Seed
     */
    decodeSeed(seed) {
        const decoded = hi_base32_1.default.decode.asBytes(seed);
        return {
            seed: Buffer.from(decoded),
        };
    }
    /**
     * keyPairFromSeed generates an object with secretKey and publicKey using the substrate sdk
     * @param seed 32 bytes long seed
     * @returns KeyPair
     */
    keyPairFromSeed(seed) {
        const keyring = new keyring_1.Keyring({ type: 'ed25519' });
        const keyringPair = keyring.addFromSeed(seed);
        const pairJson = keyringPair.toJson();
        const decodedKeyPair = (0, decode_1.decodePair)('', (0, util_crypto_1.base64Decode)(pairJson.encoded), pairJson.encoding.type);
        return new _1.KeyPair({ prv: Buffer.from(decodedKeyPair.secretKey).toString('hex') });
    }
    /**
     * Signing function. Implement this on the OFFLINE signing device.
     *
     * @param {KeyringPair} pair - The signing pair.
     * @param {string} signingPayload - Payload to sign.
     * @param {UnsignedTransaction} transaction - raw transaction to sign
     * @param {Object} options
     * @param {HexString} options.metadataRpc - metadata that is needed for substrate to sign
     * @param {TypeRegistry} options.registry - metadata that is needed for substrate to sign
     */
    createSignedTx(pair, signingPayload, transaction, options) {
        const { registry, metadataRpc } = options;
        const { signature } = registry
            .createType('ExtrinsicPayload', signingPayload, {
            version: Extrinsic_1.EXTRINSIC_VERSION,
        })
            .sign(pair);
        // Serialize a signed transaction.
        return this.serializeSignedTransaction(transaction, signature, metadataRpc, registry);
    }
    /**
     * Serializes the signed transaction
     *
     * @param transaction Transaction to serialize
     * @param signature Signature of the message
     * @param metadataRpc Network metadata
     * @param registry Transaction registry
     * @returns string Serialized transaction
     */
    serializeSignedTransaction(transaction, signature, metadataRpc, registry) {
        return txwrapper_polkadot_1.construct.signedTx(transaction, signature, {
            metadataRpc,
            registry,
        });
    }
    /**
     * Decodes the substrate address from the given format
     *
     * @param {string} address
     * @param {number} [ss58Format]
     * @returns {string}
     */
    decodeSubstrateAddress(address, ss58Format) {
        const keypair = new _1.KeyPair({ pub: Buffer.from((0, keyring_1.decodeAddress)(address, undefined, ss58Format)).toString('hex') });
        return keypair.getAddress(ss58Format);
    }
    /**
     * Decodes the substrate address from the given format
     *
     * @param {string} address
     * @param {number} [ss58Format]
     * @returns {string}
     */
    encodeSubstrateAddress(address, ss58Format) {
        return (0, keyring_1.encodeAddress)(address, ss58Format);
    }
    /**
     * Retrieves the txHash of a signed txHex
     *
     * @param txHex signed transaction hex
     * @returns {string}
     */
    getTxHash(txHex) {
        return txwrapper_polkadot_1.construct.txHash(txHex);
    }
    isSigningPayload(payload) {
        return payload.blockHash !== undefined;
    }
    isTransfer(arg) {
        return arg.dest?.id !== undefined && arg.value !== undefined;
    }
    isTransferAll(arg) {
        return arg.dest?.id !== undefined && arg.keepAlive !== undefined;
    }
    isAddStake(arg) {
        return (arg.amountStaked !== undefined &&
            arg.hotkey !== undefined &&
            arg.netuid !== undefined);
    }
    isRemoveStake(arg) {
        return (arg.amountUnstaked !== undefined &&
            arg.hotkey !== undefined &&
            arg.netuid !== undefined);
    }
    isBond(arg) {
        return (arg.value !== undefined &&
            arg.controller !== undefined &&
            arg.payee !== undefined);
    }
    isBondExtra(arg) {
        return arg.maxAdditional !== undefined;
    }
    isNominate(arg) {
        return arg.targets !== undefined;
    }
    isChill(arg) {
        return true; // Chill has no arguments, so any object can be considered ChillArgs
    }
    isUnbond(arg) {
        return arg.value !== undefined;
    }
    isWithdrawUnbonded(arg) {
        return arg.numSlashingSpans !== undefined;
    }
    isBatch(arg) {
        return arg.calls !== undefined && Array.isArray(arg.calls);
    }
    isMoveStake(arg) {
        return (arg.originHotkey !== undefined &&
            arg.destinationHotkey !== undefined &&
            arg.originNetuid !== undefined &&
            arg.destinationNetuid !== undefined &&
            arg.alphaAmount !== undefined);
    }
    /**
     * extracts and returns the signature in hex format given a raw signed transaction
     *
     * @param {string} rawTx signed raw transaction
     * @param options registry substrate registry used to retrieve the signature
     */
    recoverSignatureFromRawTx(rawTx, options) {
        const { registry } = options;
        const methodCall = registry.createType('Extrinsic', rawTx, {
            isSigned: true,
        });
        let signature = (0, util_1.u8aToHex)(methodCall.signature);
        // remove 0x from the signature since this is how it's returned from TSS signing
        if (signature.startsWith('0x')) {
            signature = signature.substr(2);
        }
        return signature;
    }
    /**
     * Decodes the dot address from the given format
     *
     * @param {string} address
     * @param {number} [ss58Format]
     * @returns {KeyPair}
     */
    decodeSubstrateAddressToKeyPair(address, ss58Format) {
        return new _1.KeyPair({ pub: Buffer.from((0, keyring_1.decodeAddress)(address, undefined, ss58Format)).toString('hex') });
    }
    /**
     * Checks whether the given input is a hex string with with 0 value
     * used to check whether a given transaction is immortal or mortal
     * @param hexValue
     */
    isZeroHex(hexValue) {
        return hexValue === '0x00';
    }
    getMaterial(networkType) {
        throw new Error('Method not implemented.');
    }
    /**
     * Decodes a substrate transaction from raw transaction hex
     *
     * @param {string} txHex - The raw transaction hex string to decode (signed or unsigned)
     * @param {Material} material - Network material containing metadata and chain information
     * @param {boolean} [isImmortalEra] - Whether the transaction uses immortal era (optional)
     * @returns {DecodedSignedTx | DecodedSigningPayload} The decoded transaction object
     */
    decodeTransaction(txHex, material, isImmortalEra = false) {
        try {
            const registry = singletonRegistry_1.SingletonRegistry.getInstance(material);
            // Attempt to decode as a signed transaction or unsigned transaction
            const decoded = (0, txwrapper_polkadot_1.decode)(txHex, {
                metadataRpc: material.metadata,
                registry,
                isImmortalEra,
            });
            return decoded;
        }
        catch (error) {
            throw new Error(`Failed to decode transaction: ${error}`);
        }
    }
}
exports.Utils = Utils;
const utils = new Utils();
exports.default = utils;
