// @bitgo/sdk-coin-arbeth/dist/src/arbethToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbethToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class ArbethToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, ArbethToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, ArbethToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(ArbethToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /**
     * Make a query to Arbiscan for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response from Arbiscan
     */
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].arbiscanApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].arbiscanBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    getFullName() {
        return 'Arbeth Token';
    }
    supportsTss() {
        return true;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
}
exports.ArbethToken = ArbethToken;
ArbethToken.coinNames = {
    Mainnet: 'arbeth',
    Testnet: 'tarbeth',
};
