// @bitgo/sdk-coin-polygon/dist/src/lib/resources.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mainnetCommon = exports.testnetCommon = void 0;
const statics_1 = require("@bitgo/statics");
const common_1 = __importDefault(require("@ethereumjs/common"));
/**
 * A Common object defining the chain and the hardfork for Polygon Testnet
 */
exports.testnetCommon = common_1.default.custom({
    name: 'polygon amoy testnet',
    networkId: statics_1.coins.get('tpolygon').network.chainId,
    chainId: statics_1.coins.get('tpolygon').network.chainId,
}, {
    baseChain: 'sepolia',
    hardfork: 'london',
    eips: [1559],
});
/**
 * A Common object defining the chain and the hardfork for Polygon Mainnet
 */
exports.mainnetCommon = common_1.default.forCustomChain('mainnet', {
    name: 'polygon mainnet',
    networkId: statics_1.coins.get('polygon').network.chainId,
    chainId: statics_1.coins.get('polygon').network.chainId,
}, 'london');
