// @bitgo/utxo-lib/dist/src/addressFormat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addressFormats = void 0;
exports.isSupportedAddressFormat = isSupportedAddressFormat;
exports.fromOutputScriptWithFormat = fromOutputScriptWithFormat;
exports.toOutputScriptWithFormat = toOutputScriptWithFormat;
exports.toOutputScriptAndFormat = toOutputScriptAndFormat;
exports.toOutputScriptTryFormats = toOutputScriptTryFormats;
exports.toCanonicalFormat = toCanonicalFormat;
/**
 * Implements methods for nonstandard (non-canonical) address formats.
 *
 * Use `toOutputScriptTryFormats()` instead of `toOutputScript()` to parse addresses in
 * non-canonical formats
 */
const networks_1 = require("./networks");
const address_1 = require("./address");
const bitgo_1 = require("./bitgo");
exports.addressFormats = ['default', 'cashaddr'];
function isCashaddrNetwork(network) {
    return (0, networks_1.isValidNetwork)(network) && [networks_1.networks.bitcoincash, networks_1.networks.ecash].includes((0, networks_1.getMainnet)(network));
}
/**
 * @param format
 * @param network
 * @return true iff format is supported for network
 */
function isSupportedAddressFormat(format, network) {
    switch (format) {
        case 'default':
            return true;
        case 'cashaddr':
            return isCashaddrNetwork(network);
    }
    throw new Error(`unknown address format ${format}`);
}
/**
 * @param outputScript
 * @param format
 * @param network
 * @return address formatted using provided AddressFormat
 */
function fromOutputScriptWithFormat(outputScript, format, network) {
    if (!isSupportedAddressFormat(format, network)) {
        throw new Error(`unsupported address format ${format} for network ${(0, networks_1.getNetworkName)(network)}`);
    }
    return isCashaddrNetwork(network)
        ? bitgo_1.bcashAddress.fromOutputScriptWithFormat(outputScript, format, network)
        : (0, address_1.fromOutputScript)(outputScript, network);
}
/**
 * @param address
 * @param format
 * @param network
 * @return output script parsed with provided AddressFormat
 */
function toOutputScriptWithFormat(address, format, network) {
    if (!isSupportedAddressFormat(format, network)) {
        throw new Error(`unsupported address format ${format} for network ${(0, networks_1.getNetworkName)(network)}`);
    }
    return isCashaddrNetwork(network)
        ? bitgo_1.bcashAddress.toOutputScriptWithFormat(address, format, network)
        : (0, address_1.toOutputScript)(address, network);
}
/**
 * Attempts to parse address with different address formats, returns first hit.
 * @param address
 * @param network
 * @param formats - defaults to all supported address formats for network
 * @return tuple with [AddressFormat, Buffer] containing format and parsed output script
 */
function toOutputScriptAndFormat(address, network, formats) {
    if (!formats) {
        formats = exports.addressFormats.filter((f) => isSupportedAddressFormat(f, network));
    }
    for (const format of formats) {
        try {
            return [format, toOutputScriptWithFormat(address, format, network)];
        }
        catch (e) {
            // try next
        }
    }
    throw new Error(`could not parse outputScript [formats=${formats}]`);
}
/**
 * Same as `toOutputScriptAndFormat`, only returning script
 * @param address - {@see toOutputScriptAndFormat}
 * @param network - {@see toOutputScriptAndFormat}
 * @param formats - {@see toOutputScriptAndFormat}
 * @return parsed output script
 */
function toOutputScriptTryFormats(address, network, formats) {
    const [, outputScript] = toOutputScriptAndFormat(address, network, formats);
    return outputScript;
}
/**
 * @param address
 * @param network
 * @return address in canonical format
 */
function toCanonicalFormat(address, network) {
    return (0, address_1.fromOutputScript)(toOutputScriptTryFormats(address, network), network);
}
