// @bitgo/utxo-lib/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = exports.supportsSegwit = exports.supportsTaproot = exports.isTestnet = exports.isMainnet = exports.getTestnet = exports.getMainnet = exports.getNetworkName = exports.isValidNetwork = exports.getNetworkList = exports.networks = exports.p2trPayments = exports.ECPair = exports.ecc = exports.bip32 = exports.testutil = exports.taproot = exports.classify = exports.addressFormat = exports.address = exports.bitgo = exports.bufferutils = void 0;
__exportStar(require("bitcoinjs-lib"), exports);
exports.bufferutils = require("bitcoinjs-lib/src/bufferutils");
exports.bitgo = require("./bitgo");
exports.address = require("./address");
exports.addressFormat = require("./addressFormat");
exports.classify = require("./classify");
exports.taproot = require("./taproot");
exports.testutil = require("./testutil");
const secp256k1_1 = require("@bitgo/secp256k1");
Object.defineProperty(exports, "bip32", { enumerable: true, get: function () { return secp256k1_1.bip32; } });
Object.defineProperty(exports, "ecc", { enumerable: true, get: function () { return secp256k1_1.ecc; } });
Object.defineProperty(exports, "ECPair", { enumerable: true, get: function () { return secp256k1_1.ECPair; } });
exports.p2trPayments = require("./payments");
var networks_1 = require("./networks");
Object.defineProperty(exports, "networks", { enumerable: true, get: function () { return networks_1.networks; } });
Object.defineProperty(exports, "getNetworkList", { enumerable: true, get: function () { return networks_1.getNetworkList; } });
Object.defineProperty(exports, "isValidNetwork", { enumerable: true, get: function () { return networks_1.isValidNetwork; } });
Object.defineProperty(exports, "getNetworkName", { enumerable: true, get: function () { return networks_1.getNetworkName; } });
Object.defineProperty(exports, "getMainnet", { enumerable: true, get: function () { return networks_1.getMainnet; } });
Object.defineProperty(exports, "getTestnet", { enumerable: true, get: function () { return networks_1.getTestnet; } });
Object.defineProperty(exports, "isMainnet", { enumerable: true, get: function () { return networks_1.isMainnet; } });
Object.defineProperty(exports, "isTestnet", { enumerable: true, get: function () { return networks_1.isTestnet; } });
Object.defineProperty(exports, "supportsTaproot", { enumerable: true, get: function () { return networks_1.supportsTaproot; } });
Object.defineProperty(exports, "supportsSegwit", { enumerable: true, get: function () { return networks_1.supportsSegwit; } });
var transaction_builder_1 = require("./transaction_builder");
Object.defineProperty(exports, "TransactionBuilder", { enumerable: true, get: function () { return transaction_builder_1.TransactionBuilder; } });
