// @bitgo/sdk-core/dist/src/bitgo/utils/tss/ecdsa/typesMPCv2.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMPCv2KeyRequestResponse = exports.generateMPCv2KeyRequestBody = exports.MPCv2PartiesEnum = void 0;
const t = __importStar(require("io-ts"));
const public_types_1 = require("@bitgo/public-types");
var MPCv2PartiesEnum;
(function (MPCv2PartiesEnum) {
    MPCv2PartiesEnum[MPCv2PartiesEnum["USER"] = 0] = "USER";
    MPCv2PartiesEnum[MPCv2PartiesEnum["BACKUP"] = 1] = "BACKUP";
    MPCv2PartiesEnum[MPCv2PartiesEnum["BITGO"] = 2] = "BITGO";
})(MPCv2PartiesEnum || (exports.MPCv2PartiesEnum = MPCv2PartiesEnum = {}));
exports.generateMPCv2KeyRequestBody = t.union([
    public_types_1.MPCv2KeyGenRound1Request,
    public_types_1.MPCv2KeyGenRound2Request,
    public_types_1.MPCv2KeyGenRound3Request,
]);
exports.generateMPCv2KeyRequestResponse = t.union([
    public_types_1.MPCv2KeyGenRound1Response,
    public_types_1.MPCv2KeyGenRound2Response,
    public_types_1.MPCv2KeyGenRound3Response,
]);
