// @bitgo/sdk-coin-rune/dist/test/resources/rune.js
"use strict";
// Get the test data by running the scripts for the particular coin from coin-sandbox repo.
Object.defineProperty(exports, "__esModule", { value: true });
exports.mainnetGasAmounts = exports.mainnetCoinAmounts = exports.txIds = exports.blockHash = exports.mainnetAddress = exports.TEST_ACCOUNT = void 0;
exports.TEST_ACCOUNT = {
    pubAddress: 'thor1xrh89ced02ea6928kcwectgtkk7k4954dags7v',
    testnetPubAddress: 'sthor1xrh89ced02ea6928kcwectgtkk7k4954ey5xgj',
    compressedPublicKey: '0371b155f21bf4426d9adeda30c858f523da3d20ca9c2a1c27b83baeac3ed6f184',
    compressedPublicKeyTwo: '02e04ea328d98de663224a126a1c64740e341976f0360d45511c954e61fba30466',
    uncompressedPublicKey: '0471b155f21bf4426d9adeda30c858f523da3d20ca9c2a1c27b83baeac3ed6f1843a0b64237e638ba7129d2dbe86a5c5cbd1e85f6e5ba6606b6c1f98e1b5d23b87',
    privateKey: 'c9cfbf9d86f86d55ce358b15402bd4f9e274b5249ff9be1b852553a36681a798',
    extendedPrv: 'xprv9s21ZrQH143K4CFciZ9dkVK2gyAj8ag8PUPm7vVZhbUhA9mQPK9f4BhzqkDsKojLhJWQndiRKPbwFaEy6dGv6eDBZkMimpiDX8ZvNN2kCVH',
    extendedPub: 'xpub661MyMwAqRbcGgL5page7dFmF11DY3PykhKMvJuBFw1g2x6YvrTubz2Uh2Xa4qmBvtxVPxy2YCpRAQQCQQYz5MctWsCdWvDwybRJvaGubjL',
};
exports.mainnetAddress = {
    address1: 'thor1fwk9jl6kfflurj9p0wt098kxl02gle4yhnm687',
    address2: 'thor17hwqt302e5f2xm4h95ma8wuggqkvfzgvsnh5z9',
    address3: 'thor17hwqt302e5f2xm4h95ma8wuggqkvfzgvsnh5',
    address4: 'tho166n4w5039meulfa3p6ydg60ve6ueac7tlt0jws',
    validatorAddress1: 'thor1v882x92avxcmkaucm2h6cp6j2qc4lhll0vfsex',
    validatorAddress2: 'thor1nl0hc33pllze4athmvyaj9ky35le30vvhx3a25',
    validatorAddress3: 'tho140wms8h9pm5dmj832lwnhw45qvt25v0ps888sf',
    validatorAddress4: 'thor1hjpct8pd9d48vyqltaqunltwx9twm57l3qmay',
    noMemoIdAddress: 'thor1pxfnmdjc0esuswlx8uxguqhgywchhuv4k9ntna',
    validMemoIdAddress: 'thor1pxfnmdjc0esuswlx8uxguqhgywchhuv4k9ntna?memoId=2',
    invalidMemoIdAddress: 'thor1pxfnmdjc0esuswlx8uxguqhgywchhuv4k9ntna?memoId=xyz',
    multipleMemoIdAddress: 'thor1pxfnmdjc0esuswlx8uxguqhgywchhuv4k9ntna?memoId=3&memoId=12',
};
exports.blockHash = {
    hash1: '10D155BC67CD47A75BA53D5EF92A69721D56D81DEB350D3AE5F3CC7876B14F1B',
    hash2: '6F17E7913CF87E6D167767D4E4167FBFB643BFFCB2571AEA6FA22566E4FBD1B5',
};
exports.txIds = {
    hash1: '7CDB56BA820B5A93014A0382014205F3BE18BF4916FE16963B35A4E9B541FA71',
    hash2: 'F2A1A5F434D389BACE737A1A57C1797C35A6DF7450B550955F03380D07BF136A',
    hash3: '2FBB379F11573583155E6F61301F7B00050417411732BF46F6CF354520077F25',
};
exports.mainnetCoinAmounts = {
    amount1: { amount: '100000', denom: 'rune' },
    amount2: { amount: '1000000', denom: 'rune' },
    amount3: { amount: '10000000', denom: 'rune' },
    amount4: { amount: '-1', denom: 'rune' },
    amount5: { amount: '1000000000', denom: 'arune' },
};
exports.mainnetGasAmounts = {
    positiveGasAmount: { amount: '100', denom: 'rune' },
    zeroGasAmount: { amount: '0', denom: 'rune' },
    emptyGasAmount: { amount: '', denom: 'rune' },
    alphabeticGasAmount: { amount: 'xyz', denom: 'rune' },
};
