// @bitgo/abstract-utxo/dist/cjs/test/unit/descriptor/builder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const testutils = __importStar(require("@bitgo/wasm-utxo/testutils"));
const builder_1 = require("../../../src/descriptor/builder");
function getDescriptorBuilderForType(name) {
    const keys = testutils.getKeyTriple('default').map((k) => k.neutered());
    switch (name) {
        case 'Wsh2Of2':
        case 'Wsh2Of3':
            return {
                name,
                keys: keys.slice(0, name === 'Wsh2Of3' ? 3 : 2),
                path: '0/*',
            };
        case 'Wsh2Of3CltvDrop':
        case 'ShWsh2Of3CltvDrop':
            return {
                name,
                keys,
                path: '0/*',
                locktime: 1,
            };
    }
}
function toComparable(builder) {
    return {
        ...builder,
        keys: builder.keys.map((k) => k.toBase58()),
    };
}
function describeForName(n) {
    describe(`DescriptorBuilder ${n}`, () => {
        it('parses descriptor template', () => {
            const builder = getDescriptorBuilderForType(n);
            const descriptor = (0, builder_1.getDescriptorFromBuilder)(builder);
            assert.deepStrictEqual(toComparable(builder), toComparable((0, builder_1.parseDescriptor)(descriptor)));
        });
    });
}
describeForName('Wsh2Of2');
describeForName('Wsh2Of3');
describeForName('Wsh2Of3CltvDrop');
describeForName('ShWsh2Of3CltvDrop');
