// @bitgo/abstract-utxo/dist/cjs/src/transaction/fixedScript/explainPsbtWasm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.explainPsbtWasm = explainPsbtWasm;
function scriptToAddress(script) {
    return `scriptPubKey:${Buffer.from(script).toString('hex')}`;
}
function isParsedWalletOutput(output) {
    return output.scriptId !== null;
}
function isParsedExternalOutput(output) {
    return output.scriptId === null;
}
function toChangeOutput(output) {
    return {
        address: output.address ?? scriptToAddress(output.script),
        amount: output.value.toString(),
        chain: output.scriptId.chain,
        index: output.scriptId.index,
        external: false,
    };
}
function toExternalOutput(output) {
    return {
        address: output.address ?? scriptToAddress(output.script),
        amount: output.value.toString(),
        external: true,
    };
}
function explainPsbtWasm(psbt, walletXpubs, params) {
    const parsed = psbt.parseTransactionWithWalletKeys(walletXpubs, { replayProtection: params.replayProtection });
    const changeOutputs = [];
    const outputs = [];
    const parsedCustomChangeOutputs = params.customChangeWalletXpubs
        ? psbt.parseOutputsWithWalletKeys(params.customChangeWalletXpubs)
        : undefined;
    const customChangeOutputs = [];
    parsed.outputs.forEach((output, i) => {
        const parseCustomChangeOutput = parsedCustomChangeOutputs?.[i];
        if (isParsedWalletOutput(output)) {
            // This is a change output
            changeOutputs.push(toChangeOutput(output));
        }
        else if (parseCustomChangeOutput && isParsedWalletOutput(parseCustomChangeOutput)) {
            customChangeOutputs.push(toChangeOutput(parseCustomChangeOutput));
        }
        else if (isParsedExternalOutput(output)) {
            outputs.push(toExternalOutput(output));
        }
        else {
            throw new Error('Invalid output');
        }
    });
    const outputAmount = outputs.reduce((sum, output) => sum + BigInt(output.amount), BigInt(0));
    const changeAmount = changeOutputs.reduce((sum, output) => sum + BigInt(output.amount), BigInt(0));
    const customChangeAmount = customChangeOutputs.reduce((sum, output) => sum + BigInt(output.amount), BigInt(0));
    return {
        id: psbt.unsignedTxid(),
        outputAmount: outputAmount.toString(),
        changeAmount: changeAmount.toString(),
        customChangeAmount: customChangeAmount.toString(),
        outputs,
        changeOutputs,
        customChangeOutputs,
        fee: parsed.minerFee.toString(),
    };
}
