// @bitgo/abstract-utxo/dist/cjs/src/transaction/fixedScript/explainTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.explainPsbt = explainPsbt;
exports.explainLegacyTx = explainLegacyTx;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_core_1 = require("@bitgo/utxo-core");
const utxo_lib_1 = require("@bitgo/utxo-lib");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const utxocore = __importStar(require("@bitgo/utxo-core"));
const recipient_1 = require("../recipient");
const getPayGoVerificationPubkey_1 = require("../getPayGoVerificationPubkey");
const keychains_1 = require("../../keychains");
const wasmUtil_1 = require("../../wasmUtil");
const names_1 = require("../../names");
function toChangeOutput(txOutput, coinName, changeInfo) {
    if (!changeInfo) {
        return undefined;
    }
    const address = (0, recipient_1.toExtendedAddressFormat)(txOutput.script, coinName);
    const change = changeInfo.find((change) => change.address === address);
    if (!change) {
        return undefined;
    }
    return {
        address,
        amount: txOutput.value.toString(),
        chain: change.chain,
        index: change.index,
        external: false,
    };
}
function outputSum(outputs) {
    return outputs.reduce((sum, output) => sum + BigInt(output.amount), BigInt(0));
}
function explainCommon(tx, params, coinName) {
    const displayOrder = ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs'];
    const changeOutputs = [];
    const customChangeOutputs = [];
    const externalOutputs = [];
    const { changeInfo, customChangeInfo } = params;
    tx.outs.forEach((currentOutput) => {
        // Try to encode the script pubkey with an address. If it fails, try to parse it as an OP_RETURN output with the prefix.
        // If that fails, then it is an unrecognized scriptPubkey and should fail
        const currentAddress = (0, recipient_1.toExtendedAddressFormat)(currentOutput.script, coinName);
        const currentAmount = BigInt(currentOutput.value);
        const changeOutput = toChangeOutput(currentOutput, coinName, changeInfo);
        if (changeOutput) {
            changeOutputs.push(changeOutput);
            return;
        }
        const customChangeOutput = toChangeOutput(currentOutput, coinName, customChangeInfo);
        if (customChangeOutput) {
            customChangeOutputs.push(customChangeOutput);
            return;
        }
        externalOutputs.push({
            address: currentAddress,
            amount: currentAmount.toString(),
            // If changeInfo has a length greater than or equal to zero, it means that the change information
            // was provided to the function but the output was not identified as change. In this case,
            // the output is external, and we can set it as so. If changeInfo is undefined, it means we were
            // given no information about change outputs, so we can't determine anything about the output,
            // so we leave it undefined.
            external: changeInfo ? true : undefined,
        });
    });
    const outputDetails = {
        outputs: externalOutputs,
        outputAmount: outputSum(externalOutputs).toString(),
        changeOutputs,
        changeAmount: outputSum(changeOutputs).toString(),
        customChangeAmount: outputSum(customChangeOutputs).toString(),
        customChangeOutputs,
    };
    let fee;
    let locktime;
    if (params.feeInfo) {
        displayOrder.push('fee');
        fee = params.feeInfo;
    }
    if (Number.isInteger(tx.locktime) && tx.locktime > 0) {
        displayOrder.push('locktime');
        locktime = tx.locktime;
    }
    return { displayOrder, id: tx.getId(), ...outputDetails, fee, locktime };
}
function getRootWalletKeys(params) {
    if (params.pubs instanceof utxo_lib_1.bitgo.RootWalletKeys) {
        return params.pubs;
    }
    const keys = params.pubs?.map((xpub) => (0, wasmUtil_1.toUtxolibBIP32)(wasm_utxo_1.BIP32.fromBase58(xpub)));
    return keys && keys.length === 3 ? new utxo_lib_1.bitgo.RootWalletKeys(keys) : undefined;
}
function getPsbtInputSignaturesCount(psbt, params) {
    const rootWalletKeys = getRootWalletKeys(params);
    return rootWalletKeys
        ? utxo_lib_1.bitgo.getSignatureValidationArrayPsbt(psbt, rootWalletKeys).map((sv) => sv[1].filter((v) => v).length)
        : Array(psbt.data.inputs.length).fill(0);
}
function getTxInputSignaturesCount(tx, params, coinName) {
    const network = (0, names_1.getNetworkFromCoinName)(coinName);
    const prevOutputs = params.txInfo?.unspents?.map((u) => utxo_lib_1.bitgo.toOutput(u, network));
    const rootWalletKeys = getRootWalletKeys(params);
    const { unspents = [] } = params.txInfo ?? {};
    // get the number of signatures per input
    return tx.ins.map((input, idx) => {
        if (unspents.length !== tx.ins.length) {
            return 0;
        }
        if (!prevOutputs) {
            throw new Error(`invalid state`);
        }
        if (!rootWalletKeys) {
            // no pub keys or incorrect number of pub keys
            return 0;
        }
        try {
            return utxo_lib_1.bitgo.verifySignatureWithUnspent(tx, idx, unspents, rootWalletKeys).filter((v) => v).length;
        }
        catch (e) {
            // some other error occurred and we can't validate the signatures
            return 0;
        }
    });
}
function getChainAndIndexFromBip32Derivations(output) {
    const derivations = output.bip32Derivation ?? output.tapBip32Derivation ?? undefined;
    if (!derivations) {
        return undefined;
    }
    const paths = derivations.map((d) => d.path);
    if (!paths || paths.length !== 3) {
        throw new Error('expected 3 paths in bip32Derivation or tapBip32Derivation');
    }
    if (!paths.every((p) => paths[0] === p)) {
        throw new Error('expected all paths to be the same');
    }
    paths.forEach((path) => {
        if (paths[0] !== path) {
            throw new Error('Unable to get a single chain and index on the output because there are different paths for different keys');
        }
    });
    return utxolib.bitgo.getChainAndIndexFromPath(paths[0]);
}
function getChangeInfo(psbt, walletKeys) {
    let utxolibKeys;
    try {
        utxolibKeys = walletKeys
            ? walletKeys.map((k) => (0, wasmUtil_1.toUtxolibBIP32)(k))
            : utxolib.bitgo.getSortedRootNodes(psbt);
    }
    catch (e) {
        if (e instanceof utxolib.bitgo.ErrorNoMultiSigInputFound) {
            return undefined;
        }
        throw e;
    }
    return utxolib.bitgo.findWalletOutputIndices(psbt, utxolibKeys).map((i) => {
        const derivationInformation = getChainAndIndexFromBip32Derivations(psbt.data.outputs[i]);
        if (!derivationInformation) {
            throw new Error('could not find derivation information on bip32Derivation or tapBip32Derivation');
        }
        return {
            address: utxolib.address.fromOutputScript(psbt.txOutputs[i].script, psbt.network),
            external: false,
            ...derivationInformation,
        };
    });
}
/**
 * Extract PayGo address proof information from the PSBT if present
 * @returns Information about the PayGo proof, including the output index and address
 */
function getPayGoVerificationInfo(psbt, coinName) {
    let outputIndex = undefined;
    let address = undefined;
    // Check if this PSBT has any PayGo address proofs
    if (!utxocore.paygo.psbtOutputIncludesPaygoAddressProof(psbt)) {
        return undefined;
    }
    // This pulls the pubkey depending on given network
    const verificationPubkey = (0, getPayGoVerificationPubkey_1.getPayGoVerificationPubkey)(coinName);
    // find which output index that contains the PayGo proof
    outputIndex = utxocore.paygo.getPayGoAddressProofOutputIndex(psbt);
    if (outputIndex === undefined || !verificationPubkey) {
        return undefined;
    }
    const network = (0, names_1.getNetworkFromCoinName)(coinName);
    const output = psbt.txOutputs[outputIndex];
    address = utxolib.address.fromOutputScript(output.script, network);
    if (!address) {
        throw new Error(`Can not derive address ${address} Pay Go Attestation.`);
    }
    return { outputIndex, verificationPubkey };
}
/**
 * Extract the BIP322 messages and addresses from the PSBT inputs and perform
 * verification on the transaction to ensure that it meets the BIP322 requirements.
 * @returns An array of objects containing the message and address for each input,
 *          or undefined if no BIP322 messages are found.
 */
function getBip322MessageInfoAndVerify(psbt, coinName) {
    const network = (0, names_1.getNetworkFromCoinName)(coinName);
    const bip322Messages = [];
    for (let i = 0; i < psbt.data.inputs.length; i++) {
        const message = utxo_core_1.bip322.getBip322ProofMessageAtIndex(psbt, i);
        if (message) {
            const input = psbt.data.inputs[i];
            if (!input.witnessUtxo) {
                throw new Error(`Missing witnessUtxo for input index ${i}`);
            }
            const scriptPubKey = input.witnessUtxo.script;
            // Verify that the toSpend transaction can be recreated in the PSBT and is encoded correctly in the nonWitnessUtxo
            const toSpend = utxo_core_1.bip322.buildToSpendTransaction(scriptPubKey, message);
            // Verify that the toSpend transaction ID matches the input's referenced transaction ID
            if (toSpend.getId() !== utxolib.bitgo.getOutputIdForInput(psbt.txInputs[i]).txid) {
                throw new Error(`ToSpend transaction ID does not match the input at index ${i}`);
            }
            // Verify the input specifics
            if (psbt.txInputs[i].sequence !== 0) {
                throw new Error(`Unexpected sequence number at input index ${i}: ${psbt.txInputs[i].sequence}. Expected 0.`);
            }
            if (psbt.txInputs[i].index !== 0) {
                throw new Error(`Unexpected input index at position ${i}: ${psbt.txInputs[i].index}. Expected 0.`);
            }
            bip322Messages.push({
                message: message.toString('utf8'),
                address: utxolib.address.fromOutputScript(scriptPubKey, network),
            });
        }
    }
    if (bip322Messages.length > 0) {
        // If there is a BIP322 message in any input, all inputs must have one.
        if (bip322Messages.length !== psbt.data.inputs.length) {
            throw new Error('Inconsistent BIP322 messages across inputs.');
        }
        // Verify the transaction specifics for BIP322
        if (psbt.version !== 0 && psbt.version !== 2) {
            throw new Error(`Unsupported PSBT version for BIP322: ${psbt.version}. Expected 0 `);
        }
        if (psbt.data.outputs.length !== 1 ||
            psbt.txOutputs[0].script.toString('hex') !== '6a' ||
            psbt.txOutputs[0].value !== 0n) {
            throw new Error(`Invalid PSBT outputs for BIP322. Expected exactly one OP_RETURN output with zero value.`);
        }
        return bip322Messages;
    }
    return undefined;
}
/**
 * Decompose a raw psbt into useful information, such as the total amounts,
 * change amounts, and transaction outputs.
 *
 * @param psbt {bitgo.UtxoPsbt} The PSBT to explain
 * @param pubs {bitgo.RootWalletKeys | string[]} The public keys to use for the explanation
 * @param coinName {UtxoCoinName} The coin name to use for the explanation
 * @param strict {boolean} Whether to throw an error if the PayGo address proof is invalid
 */
function explainPsbt(psbt, params, coinName, { strict = true } = {}) {
    const network = (0, names_1.getNetworkFromCoinName)(coinName);
    const payGoVerificationInfo = getPayGoVerificationInfo(psbt, coinName);
    if (payGoVerificationInfo) {
        try {
            utxocore.paygo.verifyPayGoAddressProof(psbt, payGoVerificationInfo.outputIndex, Buffer.from(wasm_utxo_1.BIP32.fromBase58(payGoVerificationInfo.verificationPubkey).publicKey));
        }
        catch (e) {
            if (strict) {
                throw e;
            }
            console.error(e);
        }
    }
    const messages = getBip322MessageInfoAndVerify(psbt, coinName);
    const changeInfo = getChangeInfo(psbt);
    const customChangeInfo = params.customChangePubs
        ? getChangeInfo(psbt, (0, keychains_1.toBip32Triple)(params.customChangePubs))
        : undefined;
    const tx = psbt.getUnsignedTx();
    const common = explainCommon(tx, { ...params, changeInfo, customChangeInfo }, coinName);
    const inputSignaturesCount = getPsbtInputSignaturesCount(psbt, params);
    // Set fee from subtracting inputs from outputs
    const outputAmount = psbt.txOutputs.reduce((cumulative, curr) => cumulative + BigInt(curr.value), BigInt(0));
    const inputAmount = psbt.txInputs.reduce((cumulative, txInput, i) => {
        const data = psbt.data.inputs[i];
        if (data.witnessUtxo) {
            return cumulative + BigInt(data.witnessUtxo.value);
        }
        else if (data.nonWitnessUtxo) {
            const tx = utxo_lib_1.bitgo.createTransactionFromBuffer(data.nonWitnessUtxo, network, { amountType: 'bigint' });
            return cumulative + BigInt(tx.outs[txInput.index].value);
        }
        else {
            throw new Error('could not find value on input');
        }
    }, BigInt(0));
    return {
        ...common,
        fee: (inputAmount - outputAmount).toString(),
        inputSignatures: inputSignaturesCount,
        signatures: inputSignaturesCount.reduce((prev, curr) => (curr > prev ? curr : prev), 0),
        messages,
    };
}
function explainLegacyTx(tx, params, coinName) {
    const common = explainCommon(tx, params, coinName);
    const inputSignaturesCount = getTxInputSignaturesCount(tx, params, coinName);
    return {
        ...common,
        inputSignatures: inputSignaturesCount,
        signatures: inputSignaturesCount.reduce((prev, curr) => (curr > prev ? curr : prev), 0),
    };
}
