// @bitgo/abstract-utxo/dist/cjs/src/transaction/fixedScript/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.signTransaction = exports.verifyTransaction = exports.parseTransaction = exports.explainPsbtWasm = exports.explainLegacyTx = exports.explainPsbt = void 0;
var explainTransaction_1 = require("./explainTransaction");
Object.defineProperty(exports, "explainPsbt", { enumerable: true, get: function () { return explainTransaction_1.explainPsbt; } });
Object.defineProperty(exports, "explainLegacyTx", { enumerable: true, get: function () { return explainTransaction_1.explainLegacyTx; } });
var explainPsbtWasm_1 = require("./explainPsbtWasm");
Object.defineProperty(exports, "explainPsbtWasm", { enumerable: true, get: function () { return explainPsbtWasm_1.explainPsbtWasm; } });
var parseTransaction_1 = require("./parseTransaction");
Object.defineProperty(exports, "parseTransaction", { enumerable: true, get: function () { return parseTransaction_1.parseTransaction; } });
var verifyTransaction_1 = require("./verifyTransaction");
Object.defineProperty(exports, "verifyTransaction", { enumerable: true, get: function () { return verifyTransaction_1.verifyTransaction; } });
var signTransaction_1 = require("./signTransaction");
Object.defineProperty(exports, "signTransaction", { enumerable: true, get: function () { return signTransaction_1.signTransaction; } });
__exportStar(require("./signLegacyTransaction"), exports);
__exportStar(require("./SigningError"), exports);
__exportStar(require("./replayProtection"), exports);
