// @bitgo/abstract-utxo/dist/cjs/src/transaction/fixedScript/parseTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTransaction = parseTransaction;
const assert_1 = __importDefault(require("assert"));
const lodash_1 = __importDefault(require("lodash"));
const sdk_core_1 = require("@bitgo/sdk-core");
const keychains_1 = require("../../keychains");
const verifyKey_1 = require("../../verifyKey");
const recipient_1 = require("../recipient");
const outputDifference_1 = require("../outputDifference");
const tnumber_1 = require("../../tnumber");
const parseOutput_1 = require("./parseOutput");
function toCanonicalTransactionRecipient(coin, output) {
    const amount = BigInt(output.valueString);
    (0, recipient_1.assertValidTransactionRecipient)({ amount, address: output.address });
    (0, assert_1.default)(output.address, 'address is required');
    if ((0, recipient_1.isScriptRecipient)(output.address)) {
        return { amount, address: output.address };
    }
    return { amount, address: coin.canonicalAddress(output.address) };
}
async function parseRbfTransaction(coin, params) {
    const { txParams, wallet } = params;
    (0, assert_1.default)(txParams.rbfTxIds);
    (0, assert_1.default)(txParams.rbfTxIds.length === 1);
    const txToBeReplaced = await wallet.getTransaction({ txHash: txParams.rbfTxIds[0], includeRbf: true });
    const recipients = txToBeReplaced.outputs.flatMap((output) => {
        // For self-sends, the walletId will be the same as the wallet's id
        if (output.wallet === wallet.id()) {
            return [];
        }
        return [toCanonicalTransactionRecipient(coin, output)];
    });
    // Recurse into parseTransaction with the derived recipients and without rbfTxIds
    return parseTransaction(coin, {
        ...params,
        txParams: {
            ...txParams,
            recipients,
            rbfTxIds: undefined,
        },
    });
}
function toExpectedOutputs(coin, txParams) {
    // verify that each recipient from txParams has their own output
    const expectedOutputs = (txParams.recipients ?? []).flatMap((output) => {
        if (output.address === undefined) {
            (0, assert_1.default)('script' in output, 'script is required for non-encodeable scriptPubkeys');
            if (output.amount.toString() !== '0') {
                throw new Error(`Only zero amounts allowed for non-encodeable scriptPubkeys: ${output}`);
            }
            return [
                {
                    script: (0, recipient_1.toOutputScript)(output, coin.name),
                    value: output.amount === 'max' ? 'max' : BigInt(output.amount),
                },
            ];
        }
        return [
            {
                script: (0, recipient_1.fromExtendedAddressFormatToScript)(output.address, coin.name),
                value: output.amount === 'max' ? 'max' : BigInt(output.amount),
            },
        ];
    });
    if (txParams.allowExternalChangeAddress && txParams.changeAddress) {
        expectedOutputs.push({
            script: (0, recipient_1.toOutputScript)(txParams.changeAddress, coin.name),
            // When an external change address is explicitly specified, count all outputs going towards that
            // address in the expected outputs (regardless of the output amount)
            value: 'max',
            // Note that the change output is not required to exist, so we mark it as optional.
            optional: true,
        });
    }
    return expectedOutputs;
}
function verifyCustomChangeKeys(userKeychain, customChange) {
    for (const keyIndex of [sdk_core_1.KeyIndices.USER, sdk_core_1.KeyIndices.BACKUP, sdk_core_1.KeyIndices.BITGO]) {
        if (!(0, verifyKey_1.verifyKeySignature)({
            userKeychain,
            keychainToVerify: customChange.keys[keyIndex],
            keySignature: customChange.signatures[keyIndex],
        })) {
            throw new Error(`failed to verify custom change ${sdk_core_1.KeyIndices[keyIndex].toLowerCase()} key signature`);
        }
    }
}
async function parseTransaction(coin, params) {
    const { txParams, txPrebuild, wallet, verification = {}, reqId } = params;
    // Branch off early for RBF transactions
    if (txParams.rbfTxIds) {
        return parseRbfTransaction(coin, params);
    }
    if (!lodash_1.default.isUndefined(verification.disableNetworking) && !lodash_1.default.isBoolean(verification.disableNetworking)) {
        throw new Error('verification.disableNetworking must be a boolean');
    }
    const disableNetworking = verification.disableNetworking;
    // obtain the keychains and key signatures
    let keychains = verification.keychains;
    if (!keychains) {
        if (disableNetworking) {
            throw new Error('cannot fetch keychains without networking');
        }
        keychains = await (0, keychains_1.fetchKeychains)(coin, wallet, reqId);
    }
    if (!keychains_1.UtxoNamedKeychains.is(keychains)) {
        throw new Error('invalid keychains');
    }
    const keychainArray = (0, keychains_1.toKeychainTriple)(keychains);
    if (lodash_1.default.isUndefined(txPrebuild.txHex)) {
        throw new Error('missing required txPrebuild property txHex');
    }
    const expectedOutputs = toExpectedOutputs(coin, txParams);
    // get the keychains from the custom change wallet if needed
    let customChange;
    const { customChangeWalletId = undefined } = wallet.coinSpecific() || {};
    if (customChangeWalletId) {
        // fetch keychains from custom change wallet for deriving addresses.
        // These keychains should be signed and this should be verified in verifyTransaction
        const customChangeKeySignatures = wallet._wallet.customChangeKeySignatures;
        const customChangeWallet = await coin.wallets().get({ id: customChangeWalletId });
        const customChangeKeys = await (0, keychains_1.fetchKeychains)(coin, customChangeWallet, reqId);
        if (!customChangeKeys) {
            throw new Error('failed to fetch keychains for custom change wallet');
        }
        if (customChangeKeys.user && customChangeKeys.backup && customChangeKeys.bitgo && customChangeWallet) {
            const customChangeKeychains = [
                customChangeKeys.user,
                customChangeKeys.backup,
                customChangeKeys.bitgo,
            ];
            customChange = {
                keys: customChangeKeychains,
                signatures: [customChangeKeySignatures.user, customChangeKeySignatures.backup, customChangeKeySignatures.bitgo],
            };
        }
    }
    let customChangeXpubs;
    if (customChange) {
        verifyCustomChangeKeys(keychainArray[sdk_core_1.KeyIndices.USER], customChange);
        customChangeXpubs = (0, keychains_1.toXpubTriple)(customChange.keys);
    }
    // obtain all outputs
    const explanation = await coin.explainTransaction({
        txHex: txPrebuild.txHex,
        txInfo: txPrebuild.txInfo,
        decodeWith: txPrebuild.decodeWith,
        pubs: keychainArray.map((k) => k.pub),
        customChangeXpubs,
    });
    const allOutputs = [...explanation.outputs, ...explanation.changeOutputs];
    /**
     * Loop through all the outputs and classify each of them as either internal spends
     * or external spends by setting the "external" property to true or false on the output object.
     */
    const allOutputDetails = await Promise.all(allOutputs.map((currentOutput) => {
        return (0, parseOutput_1.parseOutput)({
            currentOutput,
            coin,
            txPrebuild,
            verification,
            keychainArray: (0, keychains_1.toKeychainTriple)(keychains),
            wallet,
            txParams: {
                recipients: txParams.recipients ?? [],
                changeAddress: txParams.changeAddress,
            },
            customChange,
            reqId,
        });
    }));
    const needsCustomChangeKeySignatureVerification = allOutputDetails.some((output) => output?.needsCustomChangeKeySignatureVerification);
    const changeOutputs = lodash_1.default.filter(allOutputDetails, { external: false });
    function toComparableOutputsWithExternal(outputs) {
        return outputs.map((output) => ({
            script: (0, recipient_1.fromExtendedAddressFormatToScript)(output.address, coin.name),
            value: output.amount === 'max' ? 'max' : BigInt(output.amount),
            external: output.external,
        }));
    }
    const missingOutputs = (0, outputDifference_1.outputDifference)(expectedOutputs, toComparableOutputsWithExternal(allOutputs));
    const implicitOutputs = (0, outputDifference_1.outputDifference)(toComparableOutputsWithExternal(allOutputDetails), expectedOutputs);
    const explicitOutputs = (0, outputDifference_1.outputDifference)(toComparableOutputsWithExternal(allOutputDetails), implicitOutputs);
    // these are all the non-wallet outputs that had been originally explicitly specified in recipients
    const explicitExternalOutputs = explicitOutputs.filter((output) => output.external);
    // this is the sum of all the originally explicitly specified non-wallet output values
    const explicitExternalSpendAmount = (0, tnumber_1.toTNumber)(explicitExternalOutputs.reduce((sum, o) => sum + BigInt(o.value), BigInt(0)), coin.amountType);
    /**
     * The calculation of the implicit external spend amount pertains to verifying the pay-as-you-go-fee BitGo
     * automatically applied to transactions sending money out of the wallet. The logic is fairly straightforward
     * in that we compare the external spend amount that was specified explicitly by the user to the portion
     * that was specified implicitly. To protect customers from people tampering with the transaction outputs, we
     * define a threshold for the maximum percentage of the implicit external spend in relation to the explicit
     * external spend.
     *
     * This has become obsolete with the intoduction of `utxocore.paygo.verifyPayGoAddressProof()`.
     */
    // make sure that all the extra addresses are change addresses
    // get all the additional external outputs the server added and calculate their values
    const implicitExternalOutputs = implicitOutputs.filter((output) => output.external);
    const implicitExternalSpendAmount = (0, tnumber_1.toTNumber)(implicitExternalOutputs.reduce((sum, o) => sum + BigInt(o.value), BigInt(0)), coin.amountType);
    function toOutputs(outputs) {
        return outputs.map((output) => ({
            address: (0, recipient_1.toExtendedAddressFormat)(output.script, coin.name),
            amount: output.value.toString(),
            external: output.external,
        }));
    }
    return {
        keychains,
        keySignatures: (0, keychains_1.getKeySignatures)(wallet) ?? {},
        outputs: allOutputDetails,
        missingOutputs: toOutputs(missingOutputs),
        explicitExternalOutputs: toOutputs(explicitExternalOutputs),
        implicitExternalOutputs: toOutputs(implicitExternalOutputs),
        changeOutputs,
        explicitExternalSpendAmount,
        implicitExternalSpendAmount,
        needsCustomChangeKeySignatureVerification,
        customChange,
    };
}
