// @bitgo/abstract-utxo/dist/cjs/src/transaction/fixedScript/SigningError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionSigningError = exports.BulkSigningError = exports.InputSigningError = void 0;
class InputSigningError extends Error {
    static expectedWalletUnspent(inputIndex, inputType, unspent) {
        return new InputSigningError(inputIndex, inputType, unspent, `not a wallet unspent, not a replay protection unspent`);
    }
    constructor(inputIndex, inputType, unspent, reason) {
        super(`signing error at input ${inputIndex}: type=${inputType} unspentId=${unspent.id}: ${reason}`);
        this.inputIndex = inputIndex;
        this.inputType = inputType;
        this.unspent = unspent;
        this.reason = reason;
    }
}
exports.InputSigningError = InputSigningError;
class BulkSigningError extends Error {
    constructor(reason) {
        const reasonMessage = reason instanceof Error ? reason.message : reason;
        super(`bulk signing error: ${reasonMessage}`);
        this.reason = reason;
    }
}
exports.BulkSigningError = BulkSigningError;
class TransactionSigningError extends Error {
    constructor(signErrors, verifyError) {
        super(`sign errors at inputs: [${signErrors.join(',')}], ` +
            `verify errors at inputs: [${verifyError.join(',')}], see log for details`);
    }
}
exports.TransactionSigningError = TransactionSigningError;
