// @bitgo/abstract-utxo/dist/cjs/src/transaction/fixedScript/signTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.signAndVerifyPsbt = signAndVerifyPsbt;
exports.signTransaction = signTransaction;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const lodash_1 = __importDefault(require("lodash"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const wasmUtil_1 = require("../../wasmUtil");
const signLegacyTransaction_1 = require("./signLegacyTransaction");
const signPsbtUtxolib_1 = require("./signPsbtUtxolib");
const signPsbtWasm_1 = require("./signPsbtWasm");
const replayProtection_1 = require("./replayProtection");
function signAndVerifyPsbt(psbt, signerKeychain, rootWalletKeys, replayProtection) {
    if (psbt instanceof utxo_lib_1.bitgo.UtxoPsbt) {
        return (0, signPsbtUtxolib_1.signAndVerifyPsbt)(psbt, (0, wasmUtil_1.toUtxolibBIP32)(signerKeychain));
    }
    (0, assert_1.default)(rootWalletKeys, 'rootWalletKeys required for wasm-utxo signing');
    (0, assert_1.default)(replayProtection, 'replayProtection required for wasm-utxo signing');
    return (0, signPsbtWasm_1.signAndVerifyPsbtWasm)(psbt, signerKeychain, rootWalletKeys, replayProtection);
}
async function signTransaction(coin, tx, signerKeychain, coinName, params) {
    let isLastSignature = false;
    if (lodash_1.default.isBoolean(params.isLastSignature)) {
        // if build is called instead of buildIncomplete, no signature placeholders are left in the sig script
        isLastSignature = params.isLastSignature;
    }
    if (tx instanceof utxo_lib_1.bitgo.UtxoPsbt) {
        const signedPsbt = await (0, signPsbtUtxolib_1.signPsbtWithMusig2ParticipantUtxolib)(coin, tx, signerKeychain ? (0, wasmUtil_1.toUtxolibBIP32)(signerKeychain) : undefined, {
            signingStep: params.signingStep,
            walletId: params.walletId,
        });
        if (isLastSignature) {
            signedPsbt.finalizeAllInputs();
            return signedPsbt.extractTransaction();
        }
        return signedPsbt;
    }
    else if (tx instanceof wasm_utxo_1.fixedScriptWallet.BitGoPsbt) {
        (0, assert_1.default)(params.pubs, 'pubs are required for fixed script signing');
        (0, assert_1.default)((0, sdk_core_1.isTriple)(params.pubs), 'pubs must be a triple');
        const rootWalletKeys = wasm_utxo_1.fixedScriptWallet.RootWalletKeys.fromXpubs(params.pubs);
        const signedPsbt = await (0, signPsbtWasm_1.signPsbtWithMusig2ParticipantWasm)(coin, tx, signerKeychain, rootWalletKeys, {
            replayProtection: {
                publicKeys: (0, replayProtection_1.getReplayProtectionPubkeys)(coinName),
            },
            signingStep: params.signingStep,
            walletId: params.walletId,
        });
        if (isLastSignature) {
            signedPsbt.finalizeAllInputs();
            return Buffer.from(signedPsbt.extractTransaction().toBytes());
        }
        return signedPsbt;
    }
    return (0, signLegacyTransaction_1.signLegacyTransaction)(tx, signerKeychain, coinName, {
        isLastSignature,
        signingStep: params.signingStep,
        txInfo: params.txInfo,
        pubs: params.pubs,
        cosignerPub: params.cosignerPub,
    });
}
