// @bitgo/utxo-core/dist/cjs/src/descriptor/Output.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isInternalOutput = isInternalOutput;
exports.isExternalOutput = isExternalOutput;
exports.getExternalFixedAmount = getExternalFixedAmount;
exports.toDerivedDescriptorWalletOutput = toDerivedDescriptorWalletOutput;
const assert_1 = __importDefault(require("assert"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const Output_1 = require("../Output");
const derive_1 = require("./derive");
function isInternalOutput(output) {
    return 'descriptor' in output && output.descriptor !== undefined;
}
function isExternalOutput(output) {
    return !isInternalOutput(output);
}
/**
 * @return the sum of the external outputs that are not 'max'
 * @param outputs
 */
function getExternalFixedAmount(outputs) {
    return (0, Output_1.getFixedOutputSum)(outputs.filter(isExternalOutput));
}
function toDerivedDescriptorWalletOutput(output, descriptorMap) {
    const descriptor = descriptorMap.get(output.descriptorName);
    if (!descriptor) {
        throw new Error(`Descriptor not found: ${output.descriptorName}`);
    }
    (0, assert_1.default)(descriptor instanceof wasm_utxo_1.Descriptor);
    const descriptorAtIndex = (0, derive_1.getDescriptorAtIndexCheckScript)(descriptor, output.descriptorIndex, output.witnessUtxo.script, output.descriptorName);
    return {
        hash: output.hash,
        index: output.index,
        witnessUtxo: output.witnessUtxo,
        descriptor: descriptorAtIndex,
    };
}
