// @bitgo/sdk-coin-tempo/dist/test/unit/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const tempo_1 = require("../../src/tempo");
const ttempo_1 = require("../../src/ttempo");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_core_1 = require("@bitgo/sdk-core");
const should = __importStar(require("should"));
describe('Tempo Coin', function () {
    let bitgo;
    let basecoin;
    const registerCoin = (name, coinClass) => {
        bitgo.safeRegister(name, (bitgo) => {
            // Create a mock statics coin
            const mockStaticsCoin = {
                name,
                fullName: name === 'tempo' ? 'Tempo' : 'Testnet Tempo',
                network: {
                    type: name === 'tempo' ? 'mainnet' : 'testnet',
                },
                features: [],
            };
            return coinClass.createInstance(bitgo, mockStaticsCoin);
        });
    };
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        registerCoin('tempo', tempo_1.Tempo);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('tempo');
    });
    it('should instantiate the coin', function () {
        const basecoin = bitgo.coin('tempo');
        basecoin.should.be.an.instanceof(tempo_1.Tempo);
    });
    it('should return the correct coin name', function () {
        basecoin.getChain().should.equal('tempo');
        basecoin.getFullName().should.equal('Tempo');
        basecoin.getBaseFactor().should.equal(1e6);
    });
    describe('Address Validation', function () {
        it('should validate plain EVM address', function () {
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f').should.equal(true);
            basecoin.isValidAddress('0x0000000000000000000000000000000000000000').should.equal(true);
        });
        it('should validate address with memoId', function () {
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=8').should.equal(true);
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=0').should.equal(true);
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=12345').should.equal(true);
        });
        it('should reject invalid addresses', function () {
            basecoin.isValidAddress('invalid').should.equal(false);
            basecoin.isValidAddress('').should.equal(false);
            basecoin.isValidAddress('0x123').should.equal(false); // Too short
        });
        it('should reject address with invalid memoId', function () {
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=abc').should.equal(false);
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=-1').should.equal(false);
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=1.5').should.equal(false);
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=01').should.equal(false); // Leading zero
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=').should.equal(false); // Empty memoId
        });
        it('should reject address with unknown query parameters', function () {
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?invalid=123').should.equal(false);
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?foo=bar').should.equal(false);
        });
        it('should reject address with multiple memoId parameters', function () {
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=1&memoId=2').should.equal(false);
        });
        it('should reject address with extra query parameters besides memoId', function () {
            basecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=1&foo=bar').should.equal(false);
        });
    });
    describe('getAddressDetails', function () {
        it('should get address details without memoId', function () {
            const addressDetails = basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f');
            addressDetails.address.should.equal('0x2476602c78e9a5e0563320c78878faa3952b256f');
            addressDetails.baseAddress.should.equal('0x2476602c78e9a5e0563320c78878faa3952b256f');
            should.not.exist(addressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const addressDetails = basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=8');
            addressDetails.address.should.equal('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=8');
            addressDetails.baseAddress.should.equal('0x2476602c78e9a5e0563320c78878faa3952b256f');
            addressDetails.memoId.should.equal('8');
        });
        it('should throw on invalid memoId address', function () {
            (() => basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=abc')).should.throw(sdk_core_1.InvalidMemoIdError);
        });
        it('should throw on multiple memoId address', function () {
            (() => basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=1&memoId=2')).should.throw(sdk_core_1.InvalidAddressError);
        });
        it('should throw on unknown query parameters', function () {
            (() => basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f?invalid=8')).should.throw(sdk_core_1.InvalidAddressError);
        });
        it('should throw on empty memoId', function () {
            (() => basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=')).should.throw(sdk_core_1.InvalidAddressError);
        });
        it('should throw on extra query parameters besides memoId', function () {
            (() => basecoin.getAddressDetails('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=1&foo=bar')).should.throw(sdk_core_1.InvalidAddressError);
        });
    });
    describe('isValidMemoId', function () {
        it('should validate correct memoIds', function () {
            basecoin.isValidMemoId('0').should.equal(true);
            basecoin.isValidMemoId('1').should.equal(true);
            basecoin.isValidMemoId('12345').should.equal(true);
            basecoin.isValidMemoId('999999999999').should.equal(true);
        });
        it('should reject invalid memoIds', function () {
            basecoin.isValidMemoId('').should.equal(false);
            basecoin.isValidMemoId('-1').should.equal(false);
            basecoin.isValidMemoId('1.5').should.equal(false);
            basecoin.isValidMemoId('abc').should.equal(false);
            basecoin.isValidMemoId('01').should.equal(false); // Leading zero
            basecoin.isValidMemoId('00').should.equal(false);
        });
    });
    describe('isWalletAddress', function () {
        const baseAddress = '0x2476602c78e9a5e0563320c78878faa3952b256f';
        it('should verify a plain address matching the base address', async function () {
            const result = await basecoin.isWalletAddress({
                address: baseAddress,
                baseAddress,
                coinSpecific: {},
            });
            result.should.equal(true);
        });
        it('should verify a memoId address whose base matches the wallet base address', async function () {
            const result = await basecoin.isWalletAddress({
                address: `${baseAddress}?memoId=8`,
                baseAddress,
                coinSpecific: {},
            });
            result.should.equal(true);
        });
        it('should verify using rootAddress when baseAddress is not provided', async function () {
            const result = await basecoin.isWalletAddress({
                address: `${baseAddress}?memoId=42`,
                rootAddress: baseAddress,
                coinSpecific: {},
            });
            result.should.equal(true);
        });
        it('should reject address with mismatched base address', async function () {
            await basecoin
                .isWalletAddress({
                address: '0x0000000000000000000000000000000000000001?memoId=8',
                baseAddress,
                coinSpecific: {},
            })
                .should.be.rejectedWith(sdk_core_1.UnexpectedAddressError);
        });
        it('should reject an invalid address', async function () {
            await basecoin
                .isWalletAddress({
                address: 'not-a-valid-address',
                baseAddress,
                coinSpecific: {},
            })
                .should.be.rejectedWith(sdk_core_1.InvalidAddressError);
        });
        it('should not require forwarderVersion (no forwarder contracts)', async function () {
            const result = await basecoin.isWalletAddress({
                address: `${baseAddress}?memoId=100`,
                baseAddress,
                coinSpecific: { forwarderVersion: undefined },
            });
            result.should.equal(true);
        });
    });
    describe('Testnet', function () {
        let testnetBasecoin;
        before(function () {
            registerCoin('ttempo', ttempo_1.Ttempo);
            testnetBasecoin = bitgo.coin('ttempo');
        });
        it('should instantiate the testnet coin', function () {
            testnetBasecoin.should.be.an.instanceof(ttempo_1.Ttempo);
        });
        it('should return the correct testnet coin name', function () {
            testnetBasecoin.getChain().should.equal('ttempo');
            testnetBasecoin.getFullName().should.equal('Testnet Tempo');
            testnetBasecoin.getBaseFactor().should.equal(1e6);
        });
        it('should validate address with memoId on testnet', function () {
            testnetBasecoin.isValidAddress('0x2476602c78e9a5e0563320c78878faa3952b256f?memoId=8').should.equal(true);
        });
    });
});
