// @bitgo/sdk-coin-tempo/dist/test/unit/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const mocha_1 = require("mocha");
const ethers_1 = require("ethers");
const transactionBuilder_1 = require("../../src/lib/transactionBuilder");
const transaction_1 = require("../../src/lib/transaction");
const utils_1 = require("../../src/lib/utils");
const constants_1 = require("../../src/lib/constants");
const statics_1 = require("@bitgo/statics");
const ttempo_1 = require("../../src/ttempo");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const tempo_1 = require("../resources/tempo");
const mockCoinConfig = statics_1.coins.get('ttempo');
(0, mocha_1.describe)('TIP-20 Utilities', () => {
    (0, mocha_1.describe)('amountToTip20Units', () => {
        (0, mocha_1.it)('should convert decimal amount to 6-decimal units', () => {
            assert_1.default.strictEqual((0, utils_1.amountToTip20Units)('1.5'), 1500000n);
            assert_1.default.strictEqual((0, utils_1.amountToTip20Units)('100'), 100000000n);
            assert_1.default.strictEqual((0, utils_1.amountToTip20Units)('0.000001'), 1n);
        });
        (0, mocha_1.it)('should throw error for invalid amount', () => {
            assert_1.default.throws(() => (0, utils_1.amountToTip20Units)('invalid'), /Invalid amount format/);
        });
    });
    (0, mocha_1.describe)('tip20UnitsToAmount', () => {
        (0, mocha_1.it)('should convert 6-decimal units to human-readable amount', () => {
            assert_1.default.strictEqual((0, utils_1.tip20UnitsToAmount)(1500000n), '1.5');
            assert_1.default.strictEqual((0, utils_1.tip20UnitsToAmount)(100000000n), '100.0');
            assert_1.default.strictEqual((0, utils_1.tip20UnitsToAmount)(1n), '0.000001');
        });
    });
    (0, mocha_1.describe)('stringToBytes32', () => {
        (0, mocha_1.it)('should convert string to bytes32', () => {
            const result = (0, utils_1.stringToBytes32)('12345');
            assert_1.default.strictEqual(result.length, 66);
            assert_1.default.ok(result.startsWith('0x'));
        });
        (0, mocha_1.it)('should throw error for string longer than 32 bytes', () => {
            assert_1.default.throws(() => (0, utils_1.stringToBytes32)('a'.repeat(33)), /Memo too long/);
        });
    });
    (0, mocha_1.describe)('encodeTip20TransferWithMemo', () => {
        (0, mocha_1.it)('should encode transferWithMemo call', () => {
            const to = ethers_1.ethers.utils.getAddress('0x742d35cc6634c0532925a3b844bc9e7595f0beb1');
            const encoded = (0, utils_1.encodeTip20TransferWithMemo)(to, 1500000n, '12345');
            assert_1.default.ok(encoded.startsWith('0x'));
            assert_1.default.ok(encoded.length > 10);
        });
    });
    (0, mocha_1.describe)('isValidTip20Amount', () => {
        (0, mocha_1.it)('should validate amounts correctly', () => {
            assert_1.default.strictEqual((0, utils_1.isValidTip20Amount)('1.5'), true);
            assert_1.default.strictEqual((0, utils_1.isValidTip20Amount)('invalid'), false);
            assert_1.default.strictEqual((0, utils_1.isValidTip20Amount)('-5'), false);
        });
    });
});
(0, mocha_1.describe)('TIP-20 Transaction Builder', () => {
    const mockToken = ethers_1.ethers.utils.getAddress('0x1234567890123456789012345678901234567890');
    const mockRecipient = ethers_1.ethers.utils.getAddress('0x742d35cc6634c0532925a3b844bc9e7595f0beb1');
    const mockFeeToken = ethers_1.ethers.utils.getAddress('0xabcdefabcdefabcdefabcdefabcdefabcdefabcd');
    (0, mocha_1.describe)('addOperation', () => {
        (0, mocha_1.it)('should add operation with and without memo', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder.addOperation({ token: mockToken, to: mockRecipient, amount: '100' });
            builder.addOperation({ token: mockToken, to: mockRecipient, amount: '50', memo: '1' });
            const operations = builder.getOperations();
            assert_1.default.strictEqual(operations.length, 2);
            assert_1.default.strictEqual(operations[0].memo, undefined);
            assert_1.default.strictEqual(operations[1].memo, '1');
        });
        (0, mocha_1.it)('should throw error for invalid addresses', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            assert_1.default.throws(() => builder.addOperation({ token: '0xinvalid', to: mockRecipient, amount: '100' }), /Invalid token address/);
            assert_1.default.throws(() => builder.addOperation({ token: mockToken, to: '0xinvalid', amount: '100' }), /Invalid recipient address/);
        });
        (0, mocha_1.it)('should throw error for invalid amount', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            assert_1.default.throws(() => builder.addOperation({ token: mockToken, to: mockRecipient, amount: 'invalid-amount' }), /Invalid amount/);
        });
        (0, mocha_1.it)('should throw error for invalid (non-numeric) memo', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            assert_1.default.throws(() => builder.addOperation({ token: mockToken, to: mockRecipient, amount: '100', memo: 'INV-001' }), /Invalid memo/);
        });
    });
    (0, mocha_1.describe)('feeToken', () => {
        (0, mocha_1.it)('should set and get fee token', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder.feeToken(mockFeeToken);
            assert_1.default.strictEqual(builder.getFeeToken(), mockFeeToken);
        });
        (0, mocha_1.it)('should throw error for invalid fee token address', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            assert_1.default.throws(() => builder.feeToken('invalid-address'), /Invalid fee token address/);
        });
    });
    (0, mocha_1.describe)('Transaction parameters', () => {
        (0, mocha_1.it)('should set nonce, gas, maxFeePerGas, and maxPriorityFeePerGas', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder.nonce(42).gas(500000n).maxFeePerGas(1000000000n).maxPriorityFeePerGas(500000000n);
            assert_1.default.strictEqual(builder._nonce, 42);
            assert_1.default.strictEqual(builder._gas, 500000n);
            assert_1.default.strictEqual(builder._maxFeePerGas, 1000000000n);
            assert_1.default.strictEqual(builder._maxPriorityFeePerGas, 500000000n);
        });
        (0, mocha_1.it)('should throw error for invalid nonce and gas', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            assert_1.default.throws(() => builder.nonce(-1), /Invalid nonce/);
            assert_1.default.throws(() => builder.gas(0n), /Invalid gas limit/);
        });
    });
    (0, mocha_1.describe)('Method chaining', () => {
        (0, mocha_1.it)('should support fluent interface', () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            const result = builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100', memo: '5' })
                .feeToken(mockFeeToken)
                .nonce(10)
                .gas(400000n)
                .maxFeePerGas(2000000000n)
                .maxPriorityFeePerGas(1000000000n);
            assert_1.default.strictEqual(result, builder);
        });
    });
});
(0, mocha_1.describe)('TIP-20 Constants', () => {
    (0, mocha_1.it)('should have correct decimal places', () => {
        assert_1.default.strictEqual(constants_1.TIP20_DECIMALS, 6);
    });
});
(0, mocha_1.describe)('TIP-20 Transaction Build', () => {
    const mockToken = ethers_1.ethers.utils.getAddress(tempo_1.TESTNET_TOKENS.alphaUSD.address);
    const mockRecipient = ethers_1.ethers.utils.getAddress(tempo_1.TEST_RECIPIENT_ADDRESS);
    const mockFeeToken = ethers_1.ethers.utils.getAddress(tempo_1.TESTNET_TOKENS.betaUSD.address);
    (0, mocha_1.describe)('Build Transaction', () => {
        (0, mocha_1.it)('should build single-operation transaction', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100', memo: '1' })
                .feeToken(mockFeeToken)
                .nonce(0)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            assert_1.default.ok(tx instanceof transaction_1.Tip20Transaction);
            assert_1.default.strictEqual(tx.getOperationCount(), 1);
            assert_1.default.strictEqual(tx.getFeeToken(), mockFeeToken);
        });
        (0, mocha_1.it)('should build batch transaction with multiple operations', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: tempo_1.TESTNET_TOKENS.alphaUSD.address, to: mockRecipient, amount: '10', memo: '1' })
                .addOperation({ token: tempo_1.TESTNET_TOKENS.betaUSD.address, to: mockRecipient, amount: '20', memo: '2' })
                .addOperation({ token: tempo_1.TESTNET_TOKENS.thetaUSD.address, to: mockRecipient, amount: '30', memo: '3' })
                .feeToken(mockFeeToken)
                .nonce(42)
                .gas(300000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            assert_1.default.ok(tx instanceof transaction_1.Tip20Transaction);
            assert_1.default.strictEqual(tx.getOperationCount(), 3);
            assert_1.default.strictEqual(tx.isBatch(), true);
        });
        (0, mocha_1.it)('should build transaction without fee token', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '50' })
                .nonce(1)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            assert_1.default.ok(tx instanceof transaction_1.Tip20Transaction);
            assert_1.default.strictEqual(tx.getFeeToken(), undefined);
        });
    });
    (0, mocha_1.describe)('Build Validation Errors', () => {
        (0, mocha_1.it)('should throw error when no operations added', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder.nonce(0).gas(100000n).maxFeePerGas(1000000000n).maxPriorityFeePerGas(500000000n);
            await assert_1.default.rejects(async () => await builder.build(), tempo_1.ERROR_MESSAGES.noOperations);
        });
        (0, mocha_1.it)('should throw error when nonce not set', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100' })
                .gas(100000n)
                .maxFeePerGas(1000000000n)
                .maxPriorityFeePerGas(500000000n);
            await assert_1.default.rejects(async () => await builder.build(), tempo_1.ERROR_MESSAGES.missingNonce);
        });
        (0, mocha_1.it)('should throw error when gas not set', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100' })
                .nonce(0)
                .maxFeePerGas(1000000000n)
                .maxPriorityFeePerGas(500000000n);
            await assert_1.default.rejects(async () => await builder.build(), tempo_1.ERROR_MESSAGES.missingGas);
        });
        (0, mocha_1.it)('should throw error when maxFeePerGas not set', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100' })
                .nonce(0)
                .gas(100000n)
                .maxPriorityFeePerGas(500000000n);
            await assert_1.default.rejects(async () => await builder.build(), tempo_1.ERROR_MESSAGES.missingMaxFeePerGas);
        });
        (0, mocha_1.it)('should throw error when maxPriorityFeePerGas not set', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100' })
                .nonce(0)
                .gas(100000n)
                .maxFeePerGas(1000000000n);
            await assert_1.default.rejects(async () => await builder.build(), tempo_1.ERROR_MESSAGES.missingMaxPriorityFeePerGas);
        });
    });
    (0, mocha_1.describe)('Round-Trip: Build -> Serialize -> Operations Check', () => {
        (0, mocha_1.it)('should preserve operation data through build', async () => {
            const operations = [{ token: mockToken, to: mockRecipient, amount: '123', memo: '1' }];
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation(operations[0])
                .feeToken(mockFeeToken)
                .nonce(10)
                .gas(150000n)
                .maxFeePerGas(2000000000n)
                .maxPriorityFeePerGas(1000000000n);
            const tx = (await builder.build());
            const txOps = tx.getOperations();
            assert_1.default.strictEqual(txOps.length, 1);
            assert_1.default.strictEqual(txOps[0].token, operations[0].token);
            assert_1.default.strictEqual(txOps[0].to, operations[0].to);
            assert_1.default.strictEqual(txOps[0].amount, operations[0].amount);
            assert_1.default.strictEqual(txOps[0].memo, operations[0].memo);
        });
    });
    (0, mocha_1.describe)('Build -> Serialize -> Sign Flow', () => {
        (0, mocha_1.it)('should produce serializable unsigned transaction', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '50', memo: '2' })
                .feeToken(mockFeeToken)
                .nonce(5)
                .gas(100000n)
                .maxFeePerGas(2000000000n)
                .maxPriorityFeePerGas(1000000000n);
            const tx = (await builder.build());
            const unsignedHex = await tx.serialize();
            assert_1.default.ok(unsignedHex.startsWith(constants_1.AA_TRANSACTION_TYPE));
            assert_1.default.ok(/^0x[0-9a-f]+$/i.test(unsignedHex));
        });
        (0, mocha_1.it)('should support full sign and broadcast flow', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '75', memo: '3' })
                .feeToken(mockFeeToken)
                .nonce(123)
                .gas(120000n)
                .maxFeePerGas(2500000000n)
                .maxPriorityFeePerGas(1200000000n);
            const tx = (await builder.build());
            const unsignedHex = await tx.serialize();
            tx.setSignature(tempo_1.SIGNATURE_TEST_DATA.validSignature);
            const broadcastHex = await tx.toBroadcastFormat();
            assert_1.default.ok(broadcastHex.startsWith(constants_1.AA_TRANSACTION_TYPE));
            assert_1.default.ok(broadcastHex.length > unsignedHex.length, 'Signed should be longer');
        });
    });
    (0, mocha_1.describe)('Transaction JSON Representation', () => {
        (0, mocha_1.it)('should produce consistent JSON from built transaction', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100', memo: '4' })
                .feeToken(mockFeeToken)
                .nonce(50)
                .gas(200000n)
                .maxFeePerGas(3000000000n)
                .maxPriorityFeePerGas(1500000000n);
            const tx = (await builder.build());
            const json = tx.toJson();
            assert_1.default.strictEqual(json.type, constants_1.AA_TRANSACTION_TYPE);
            assert_1.default.strictEqual(json.nonce, 50);
            assert_1.default.strictEqual(json.gas, '200000');
            assert_1.default.strictEqual(json.maxFeePerGas, '3000000000');
            assert_1.default.strictEqual(json.maxPriorityFeePerGas, '1500000000');
            assert_1.default.strictEqual(json.callCount, 1);
            assert_1.default.strictEqual(json.feeToken, mockFeeToken);
        });
    });
    (0, mocha_1.describe)('Transaction id getter', () => {
        (0, mocha_1.it)('unsigned and signed transactions should have different ids', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '10' })
                .nonce(1)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            const unsignedId = tx.id;
            tx.setSignature(tempo_1.SIGNATURE_TEST_DATA.validSignature);
            const signedId = tx.id;
            assert_1.default.notStrictEqual(unsignedId, signedId, 'Signed and unsigned tx should have different ids');
        });
    });
    (0, mocha_1.describe)('outputs / inputs population', () => {
        (0, mocha_1.it)('should expose outputs with base-unit value and token address as coin', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '1.5', memo: '7' })
                .nonce(0)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            assert_1.default.strictEqual(tx.outputs.length, 1);
            assert_1.default.strictEqual(tx.outputs[0].address, mockRecipient);
            // 1.5 tokens * 10^6 = 1_500_000 base units
            assert_1.default.strictEqual(tx.outputs[0].value, '1500000');
            // coin is the token contract address, not the chain name
            assert_1.default.strictEqual(tx.outputs[0].coin?.toLowerCase(), mockToken.toLowerCase());
        });
        (0, mocha_1.it)('should expose a single input with the total in base units', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '1.5' })
                .addOperation({ token: mockToken, to: mockRecipient, amount: '3.5' })
                .nonce(0)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            assert_1.default.strictEqual(tx.inputs.length, 1);
            assert_1.default.strictEqual(tx.inputs[0].address, '');
            // (1.5 + 3.5) tokens * 10^6 = 5_000_000 base units
            assert_1.default.strictEqual(tx.inputs[0].value, '5000000');
        });
    });
    (0, mocha_1.describe)('Round-Trip: Build -> Serialize -> From (deserialization)', () => {
        (0, mocha_1.it)('should round-trip a single-operation transaction without a signature', async () => {
            const operation = { token: mockToken, to: mockRecipient, amount: '25.5', memo: '12345' };
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation(operation)
                .feeToken(mockFeeToken)
                .nonce(7)
                .gas(150000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const originalTx = (await builder.build());
            const serialized = await originalTx.serialize();
            // Deserialize via from()
            const builder2 = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder2.from(serialized);
            const restoredTx = (await builder2.build());
            const ops = restoredTx.getOperations();
            assert_1.default.strictEqual(ops.length, 1);
            assert_1.default.strictEqual(ops[0].token.toLowerCase(), operation.token.toLowerCase());
            assert_1.default.strictEqual(ops[0].to.toLowerCase(), operation.to.toLowerCase());
            assert_1.default.strictEqual(ops[0].amount, operation.amount);
            assert_1.default.strictEqual(ops[0].memo, operation.memo);
            assert_1.default.strictEqual(restoredTx.getFeeToken()?.toLowerCase(), mockFeeToken.toLowerCase());
        });
        (0, mocha_1.it)('should round-trip a batch transaction', async () => {
            // Use amounts that match tip20UnitsToAmount output: ethers formatUnits always includes decimal
            const operations = [
                { token: tempo_1.TESTNET_TOKENS.alphaUSD.address, to: mockRecipient, amount: '10.0', memo: '1' },
                { token: tempo_1.TESTNET_TOKENS.betaUSD.address, to: mockRecipient, amount: '20.0', memo: '2' },
            ];
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation(operations[0])
                .addOperation(operations[1])
                .nonce(42)
                .gas(250000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const originalTx = (await builder.build());
            const serialized = await originalTx.serialize();
            const builder2 = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder2.from(serialized);
            const restoredTx = (await builder2.build());
            assert_1.default.strictEqual(restoredTx.getOperationCount(), 2);
            assert_1.default.ok(restoredTx.isBatch());
            const json = restoredTx.toJson();
            assert_1.default.strictEqual(json.nonce, 42);
            assert_1.default.strictEqual(json.gas, '250000');
            assert_1.default.strictEqual(json.maxFeePerGas, tempo_1.TX_PARAMS.defaultMaxFeePerGas.toString());
            assert_1.default.strictEqual(json.maxPriorityFeePerGas, tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas.toString());
            const ops = restoredTx.getOperations();
            for (let i = 0; i < operations.length; i++) {
                assert_1.default.strictEqual(ops[i].token.toLowerCase(), operations[i].token.toLowerCase());
                assert_1.default.strictEqual(ops[i].to.toLowerCase(), mockRecipient.toLowerCase());
                assert_1.default.strictEqual(ops[i].amount, operations[i].amount);
                assert_1.default.strictEqual(ops[i].memo, operations[i].memo);
            }
        });
        (0, mocha_1.it)('should round-trip a signed transaction and preserve the signature', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '5' })
                .nonce(99)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const tx = (await builder.build());
            tx.setSignature(tempo_1.SIGNATURE_TEST_DATA.validSignature);
            const signedHex = await tx.toBroadcastFormat();
            // Deserialize the signed transaction
            const builder2 = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder2.from(signedHex);
            const restoredTx = (await builder2.build());
            const sig = restoredTx.getSignature();
            assert_1.default.ok(sig !== undefined, 'Signature should be preserved');
            assert_1.default.strictEqual(sig.yParity, tempo_1.SIGNATURE_TEST_DATA.validSignature.yParity);
        });
        (0, mocha_1.it)('should produce the same tx id after a serialization round-trip', async () => {
            const builder = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder
                .addOperation({ token: mockToken, to: mockRecipient, amount: '100.0', memo: '99' })
                .nonce(3)
                .gas(100000n)
                .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
                .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
            const originalTx = (await builder.build());
            const serialized = await originalTx.serialize();
            const originalId = originalTx.id;
            const builder2 = new transactionBuilder_1.Tip20TransactionBuilder(mockCoinConfig);
            builder2.from(serialized);
            const restoredTx = (await builder2.build());
            assert_1.default.strictEqual(restoredTx.id, originalId);
            const originalJson = originalTx.toJson();
            const restoredJson = restoredTx.toJson();
            assert_1.default.strictEqual(restoredJson.nonce, originalJson.nonce);
            assert_1.default.strictEqual(restoredJson.gas, originalJson.gas);
            assert_1.default.strictEqual(restoredJson.maxFeePerGas, originalJson.maxFeePerGas);
            assert_1.default.strictEqual(restoredJson.maxPriorityFeePerGas, originalJson.maxPriorityFeePerGas);
            assert_1.default.strictEqual(restoredJson.feeToken, originalJson.feeToken);
            assert_1.default.strictEqual(restoredJson.callCount, originalJson.callCount);
            const originalOps = originalTx.getOperations();
            const restoredOps = restoredTx.getOperations();
            assert_1.default.strictEqual(restoredOps.length, originalOps.length);
            assert_1.default.strictEqual(restoredOps[0].token.toLowerCase(), originalOps[0].token.toLowerCase());
            assert_1.default.strictEqual(restoredOps[0].to.toLowerCase(), originalOps[0].to.toLowerCase());
            assert_1.default.strictEqual(restoredOps[0].amount, originalOps[0].amount);
            assert_1.default.strictEqual(restoredOps[0].memo, originalOps[0].memo);
            const restoredSerialized = await restoredTx.serialize();
            assert_1.default.strictEqual(restoredSerialized, serialized);
        });
    });
});
(0, mocha_1.describe)('Tempo coin - parseTransaction / verifyTransaction', () => {
    let bitgo;
    let coin;
    const mockToken = ethers_1.ethers.utils.getAddress(tempo_1.TESTNET_TOKENS.alphaUSD.address);
    const mockRecipient = ethers_1.ethers.utils.getAddress(tempo_1.TEST_RECIPIENT_ADDRESS);
    (0, mocha_1.before)(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('ttempo', (bg) => {
            const mockStaticsCoin = {
                name: 'ttempo',
                fullName: 'Testnet Tempo',
                network: { type: 'testnet' },
                features: [],
            };
            return ttempo_1.Ttempo.createInstance(bg, mockStaticsCoin);
        });
        bitgo.initializeTestVars();
        coin = bitgo.coin('ttempo');
    });
    async function buildSerializedTx(operations, nonce = 0) {
        const builder = new transactionBuilder_1.Tip20TransactionBuilder(statics_1.coins.get('ttempo'));
        for (const op of operations) {
            builder.addOperation(op);
        }
        builder
            .nonce(nonce)
            .gas(100000n)
            .maxFeePerGas(tempo_1.TX_PARAMS.defaultMaxFeePerGas)
            .maxPriorityFeePerGas(tempo_1.TX_PARAMS.defaultMaxPriorityFeePerGas);
        const tx = (await builder.build());
        return tx.serialize();
    }
    (0, mocha_1.describe)('parseTransaction', () => {
        (0, mocha_1.it)('should parse a single-operation transaction into outputs', async () => {
            const txHex = await buildSerializedTx([{ token: mockToken, to: mockRecipient, amount: '50', memo: '1' }]);
            const parsed = await coin.parseTransaction({ txHex });
            assert_1.default.ok(Array.isArray(parsed.outputs), 'outputs should be an array');
            assert_1.default.strictEqual(parsed.outputs.length, 1);
            assert_1.default.strictEqual(parsed.outputs[0].address.toLowerCase(), mockRecipient.toLowerCase());
            // 50 tokens * 10^6 = 50_000_000 base units
            assert_1.default.strictEqual(parsed.outputs[0].amount, '50000000');
        });
        (0, mocha_1.it)('should return empty object when no txHex is provided', async () => {
            const parsed = await coin.parseTransaction({});
            assert_1.default.deepStrictEqual(parsed, {});
        });
    });
    (0, mocha_1.describe)('verifyTransaction', () => {
        (0, mocha_1.it)('should return true when no recipients are specified', async () => {
            const txHex = await buildSerializedTx([{ token: mockToken, to: mockRecipient, amount: '10' }]);
            const result = await coin.verifyTransaction({
                txPrebuild: { txHex },
                txParams: {},
            });
            assert_1.default.strictEqual(result, true);
        });
        (0, mocha_1.it)('should return true when recipients match operations', async () => {
            const txHex = await buildSerializedTx([{ token: mockToken, to: mockRecipient, amount: '100' }]);
            const result = await coin.verifyTransaction({
                txPrebuild: { txHex },
                txParams: {
                    recipients: [{ address: mockRecipient, amount: '100000000' }], // 100 * 10^6 base units
                },
            });
            assert_1.default.strictEqual(result, true);
        });
        (0, mocha_1.it)('should throw when recipient address does not match', async () => {
            const txHex = await buildSerializedTx([{ token: mockToken, to: mockRecipient, amount: '100' }]);
            const wrongAddress = ethers_1.ethers.utils.getAddress('0x1111111111111111111111111111111111111111');
            await assert_1.default.rejects(() => coin.verifyTransaction({
                txPrebuild: { txHex },
                txParams: { recipients: [{ address: wrongAddress, amount: '100000000' }] },
            }), /recipient mismatch/);
        });
        (0, mocha_1.it)('should throw when recipient amount does not match', async () => {
            const txHex = await buildSerializedTx([{ token: mockToken, to: mockRecipient, amount: '100' }]);
            await assert_1.default.rejects(() => coin.verifyTransaction({
                txPrebuild: { txHex },
                txParams: { recipients: [{ address: mockRecipient, amount: '999' }] },
            }), /amount mismatch/);
        });
        (0, mocha_1.it)('should throw when operation count differs from recipient count', async () => {
            const txHex = await buildSerializedTx([{ token: mockToken, to: mockRecipient, amount: '10' }]);
            await assert_1.default.rejects(() => coin.verifyTransaction({
                txPrebuild: { txHex },
                txParams: {
                    recipients: [
                        { address: mockRecipient, amount: '10000000' },
                        { address: mockRecipient, amount: '10000000' },
                    ],
                },
            }), /operation\(s\)/);
        });
        (0, mocha_1.it)('should return true when no txHex is provided', async () => {
            const result = await coin.verifyTransaction({ txPrebuild: {}, txParams: {} });
            assert_1.default.strictEqual(result, true);
        });
    });
});
