// @bitgo/public-types/dist/src/schema/tss/tssSettings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TssSettings = exports.TssCoinSetting = exports.TssWalletCreationSettings = void 0;
const t = __importStar(require("io-ts"));
const wallet_1 = require("../wallet");
exports.TssWalletCreationSettings = t.partial({
    multiSigTypeVersion: wallet_1.MultisigTypeVersion,
    custodialMultiSigTypeVersion: wallet_1.MultisigTypeVersion,
    coldMultiSigTypeVersion: wallet_1.MultisigTypeVersion,
    defaultWalletVersion: t.number,
});
exports.TssCoinSetting = t.intersection([
    t.type({ walletCreationSettings: exports.TssWalletCreationSettings }),
    t.partial({
        canMigrateToMPCv2: t.boolean,
    }),
]);
exports.TssSettings = t.type({
    coinSettings: t.record(t.string, exports.TssCoinSetting),
});
