// @bitgo/sdk-coin-sui/dist/src/lib/mystenlab/framework/sui-system-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUI_STAKING_POOL_SPLIT_FUN_NAME = exports.SUI_STAKING_POOL_MODULE_NAME = exports.WITHDRAW_STAKE_FUN_NAME = exports.ADD_STAKE_LOCKED_COIN_FUN_NAME = exports.ADD_STAKE_FUN_NAME = exports.SUI_SYSTEM_MODULE_NAME = exports.SUI_SYSTEM_STATE_OBJECT = exports.SUI_SYSTEM_STATE_OBJECT_ID = void 0;
const types_1 = require("../types");
/**
 * Address of the Sui System object.
 * Always the same in every Sui network (local, devnet, testnet).
 */
exports.SUI_SYSTEM_STATE_OBJECT_ID = (0, types_1.normalizeSuiObjectId)('0x5');
exports.SUI_SYSTEM_STATE_OBJECT = {
    objectId: exports.SUI_SYSTEM_STATE_OBJECT_ID,
    initialSharedVersion: 1,
    mutable: true,
};
exports.SUI_SYSTEM_MODULE_NAME = 'sui_system';
exports.ADD_STAKE_FUN_NAME = 'request_add_stake';
exports.ADD_STAKE_LOCKED_COIN_FUN_NAME = 'request_add_stake_with_locked_coin';
exports.WITHDRAW_STAKE_FUN_NAME = 'request_withdraw_stake';
// https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/staking_pool.md
exports.SUI_STAKING_POOL_MODULE_NAME = 'staking_pool';
exports.SUI_STAKING_POOL_SPLIT_FUN_NAME = 'split';
