// @bitgo/sdk-coin-avaxc/dist/test/unit/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.walletSimpleABI = void 0;
exports.decodeTransaction = decodeTransaction;
exports.parseTransaction = parseTransaction;
const ethers_1 = require("ethers");
function decodeTransaction(abi, calldata) {
    const contractInterface = new ethers_1.ethers.utils.Interface(abi);
    return contractInterface.parseTransaction({ data: calldata });
}
function parseTransaction(txHex) {
    return ethers_1.ethers.utils.parseTransaction(txHex);
}
exports.walletSimpleABI = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'toAddress',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'value',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'data',
                type: 'bytes',
            },
            {
                internalType: 'uint256',
                name: 'expireTime',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'sequenceId',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'signature',
                type: 'bytes',
            },
        ],
        name: 'sendMultiSig',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
