// bitgo/dist/test/v2/unit/tss/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nockSendSignatureShare = nockSendSignatureShare;
exports.nockGetTxRequest = nockGetTxRequest;
exports.nockGetEnterprise = nockGetEnterprise;
exports.nockGetChallenges = nockGetChallenges;
exports.nockGetChallenge = nockGetChallenge;
exports.nockGetSigningKey = nockGetSigningKey;
exports.createWalletSignatures = createWalletSignatures;
const openpgp = require("openpgp");
const sdk_core_1 = require("@bitgo/sdk-core");
const common_1 = require("../internal/tssUtils/common");
const nock = require("nock");
async function nockSendSignatureShare(params, status = 200) {
    const { signatureShare, signerShare, tssType } = params;
    const transactions = (0, common_1.getRoute)(tssType);
    const requestBody = signerShare === undefined ? { signatureShare } : { signatureShare, signerShare };
    return nock('https://bitgo.fakeurl')
        .post(`/api/v2/wallet/${params.walletId}/txrequests/${params.txRequestId + transactions}/signatureshares`, requestBody)
        .reply(status, status === 200 ? (params.response ? params.response : params.signatureShare) : { error: 'some error' });
}
async function nockGetTxRequest(params) {
    const n = nock('https://bitgo.fakeurl').get(`/api/v2/wallet/${params.walletId}/txrequests?txRequestIds=${params.txRequestId}&latest=true`);
    if (params.times !== undefined) {
        return n.times(params.times).reply(200, params.response);
    }
    return n.reply(200, params.response);
}
async function nockGetEnterprise(params) {
    const n = nock('https://bitgo.fakeurl').get(`/api/v1/enterprise/${params.enterpriseId}`);
    if (params.times !== undefined) {
        return n.times(params.times).reply(200, params.response);
    }
    return n.reply(200, params.response);
}
async function nockGetChallenges(params) {
    return nock('https://bitgo.fakeurl')
        .get('/api/v2/wallet/' + params.walletId + '/challenges')
        .reply(200, params.response);
}
async function nockGetChallenge(params) {
    const n = nock('https://bitgo.fakeurl').post('/api/v2/wallet/' + params.walletId + '/txrequests/' + params.txRequestId + params.addendum + '/challenge');
    if (params.times !== undefined) {
        return n.times(params.times).reply(200, params.response);
    }
    return n.reply(200, params.response);
}
async function nockGetSigningKey(params) {
    const path = `/api/v2/enterprise/${params.enterpriseId}/user/${params.userId}/signingkey`;
    const n = nock('https://bitgo.fakeurl').get(path);
    if (params.times !== undefined) {
        return n.times(params.times).reply(200, params.response);
    }
    return n.reply(200, params.response);
}
async function createWalletSignatures(privateKeyArmored, publicKeyToCertArmoredUser, publicKeyToCertArmoredBackup, notations) {
    const userWalletSigArmored = await (0, sdk_core_1.createSharedDataProof)(privateKeyArmored, publicKeyToCertArmoredUser, notations);
    const backupWalletSigArmored = await (0, sdk_core_1.createSharedDataProof)(privateKeyArmored, publicKeyToCertArmoredBackup, notations);
    const certsUserKey = await openpgp.readKey({ armoredKey: userWalletSigArmored });
    const certsBackupKey = await openpgp.readKey({ armoredKey: backupWalletSigArmored });
    const mergedWalletKeys = new openpgp.PacketList();
    certsUserKey.toPacketList().forEach((packet) => mergedWalletKeys.push(packet));
    certsBackupKey.toPacketList().forEach((packet) => mergedWalletKeys.push(packet));
    // the underlying function only requires two arguments but the according .d.ts file for openpgp has the further
    // arguments marked as mandatory as well.
    // Once the following PR has been merged and released we no longer need the ts-ignore:
    // https://github.com/openpgpjs/openpgpjs/pull/1576
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return openpgp.armor(openpgp.enums.armor.publicKey, mergedWalletKeys.write());
}
