// @bitgo/wasm-utxo/dist/cjs/js/descriptorWallet/DescriptorOutput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isInternalOutput = isInternalOutput;
exports.isExternalOutput = isExternalOutput;
exports.getExternalFixedAmount = getExternalFixedAmount;
exports.toDerivedDescriptorWalletOutput = toDerivedDescriptorWalletOutput;
/**
 * Descriptor output types and utilities.
 * Moved from @bitgo/utxo-core.
 */
const index_js_1 = require("../index.js");
const Output_js_1 = require("./Output.js");
const derive_js_1 = require("./derive.js");
function isInternalOutput(output) {
    return "descriptor" in output && output.descriptor !== undefined;
}
function isExternalOutput(output) {
    return !isInternalOutput(output);
}
/**
 * @return the sum of the external outputs that are not 'max'
 * @param outputs
 */
function getExternalFixedAmount(outputs) {
    return (0, Output_js_1.getFixedOutputSum)(outputs.filter(isExternalOutput));
}
function toDerivedDescriptorWalletOutput(output, descriptorMap) {
    const descriptor = descriptorMap.get(output.descriptorName);
    if (!descriptor) {
        throw new Error(`Descriptor not found: ${output.descriptorName}`);
    }
    if (!(descriptor instanceof index_js_1.Descriptor)) {
        throw new Error(`Expected Descriptor instance for ${output.descriptorName}`);
    }
    const descriptorAtIndex = (0, derive_js_1.getDescriptorAtIndexCheckScript)(descriptor, output.descriptorIndex, output.witnessUtxo.script, output.descriptorName);
    return {
        hash: output.hash,
        index: output.index,
        witnessUtxo: output.witnessUtxo,
        descriptor: descriptorAtIndex,
    };
}
