// @bitgo/account-lib/dist/test/resources/testTransactionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestTransactionBuilder = void 0;
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
/**
 * The purpose of this coin is a mock to use for the test runner since there is no easy way to mock
 * an interface with sinon without providing the mandatory fields/methods.
 * Use it along with sinon by stubbing its methods.
 */
class TestTransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor() {
        super(statics_1.coins.get('ttrx'));
        this._coinConfig = statics_1.coins.get('ttrx');
    }
    displayName() {
        return 'Test';
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    validateAddress(address, addressFormat) { }
    validateValue(value) {
        return true;
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    validateKey(key) { }
    fromImplementation(rawTransaction) {
        return rawTransaction;
    }
    signImplementation(privateKey) {
        return this._transaction;
    }
    async buildImplementation() {
        return Promise.resolve(this._transaction);
    }
    get transaction() {
        return this._transaction;
    }
    set transaction(transaction) {
        this._transaction = transaction;
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    validateRawTransaction(rawTransaction) { }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    validateTransaction(transaction) { }
}
exports.TestTransactionBuilder = TestTransactionBuilder;
