// @bitgo/utxo-core/dist/cjs/test/descriptor/parse/PatternMatcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const testutil_1 = require("../../../src/testutil");
const src_1 = require("../../../src");
const PatternMatcher_1 = require("../../../src/descriptor/parse/PatternMatcher");
function key32(seed) {
    // return x-only public key from seed
    return (0, src_1.toXOnlyPublicKey)((0, testutil_1.getKey)(seed).publicKey);
}
function getUnspendableKey() {
    return '50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0';
}
function pk(b) {
    return { 'v:pk': b.toString('hex') };
}
function sortedKeys(keys) {
    return [...keys].sort((a, b) => a.compare(b));
}
function multiArgs(threshold, keys) {
    return [threshold, ...sortedKeys(keys).map((k) => k.toString('hex'))];
}
function taprootScriptOnlyFromAst(n) {
    return wasm_utxo_1.Descriptor.fromString(wasm_utxo_1.ast.formatNode({ tr: [getUnspendableKey(), n] }), 'definite');
}
class StakingDescriptorBuilder {
    constructor(userKey, providerKeys, guardianKeys, guardianThreshold, stakingTimeLock) {
        this.userKey = userKey;
        this.providerKeys = providerKeys;
        this.guardianKeys = guardianKeys;
        this.guardianThreshold = guardianThreshold;
        this.stakingTimeLock = stakingTimeLock;
    }
    getTimelockMiniscriptNode() {
        return { and_v: [pk(this.userKey), { older: this.stakingTimeLock }] };
    }
    getWithdrawalMiniscriptNode() {
        return { and_v: [pk(this.userKey), { multi_a: multiArgs(this.guardianThreshold, this.guardianKeys) }] };
    }
    getPenaltyMiniscriptNode() {
        return {
            and_v: [
                {
                    and_v: [
                        pk(this.userKey),
                        this.providerKeys.length === 1
                            ? { 'v:pk': this.providerKeys[0].toString('hex') }
                            : { 'v:multi_a': multiArgs(1, this.providerKeys) },
                    ],
                },
                { multi_a: multiArgs(this.guardianThreshold, this.guardianKeys) },
            ],
        };
    }
    getStakingDescriptor() {
        return taprootScriptOnlyFromAst([
            this.getPenaltyMiniscriptNode(),
            [this.getWithdrawalMiniscriptNode(), this.getTimelockMiniscriptNode()],
        ]);
    }
}
// Inverse function to parse the descriptor
function parseStakingDescriptor(descriptor) {
    const pattern = {
        tr: [getUnspendableKey(), [{ $var: 'penaltyNode' }, [{ $var: 'withdrawalNode' }, { $var: 'timelockNode' }]]],
    };
    const matcher = new PatternMatcher_1.PatternMatcher();
    const descriptorNode = wasm_utxo_1.ast.fromDescriptor(descriptor);
    const result = matcher.match(descriptorNode, pattern);
    if (!result) {
        return null;
    }
    return {
        penaltyNode: result.penaltyNode,
        withdrawalNode: result.withdrawalNode,
        timelockNode: result.timelockNode,
    };
}
describe('PatternMatcher', function () {
    it('should match basic object', function () {
        const pattern = { a: { $var: 'x' }, b: 'hello' };
        const node = { a: 123, b: 'hello' };
        const vars = new PatternMatcher_1.PatternMatcher().match(node, pattern);
        assert_1.default.deepStrictEqual(vars, { x: 123 });
    });
    it('should fail on non-matching object', function () {
        const pattern = { a: { $var: 'x' }, b: 'world' };
        const node = { a: 123, b: 'hello' };
        const vars = new PatternMatcher_1.PatternMatcher().match(node, pattern);
        assert_1.default.strictEqual(vars, null);
    });
    it('should match with repeated var', function () {
        const pattern = { a: { $var: 'x' }, b: { $var: 'x' } };
        const node = { a: 123, b: 123 };
        const vars = new PatternMatcher_1.PatternMatcher().match(node, pattern);
        assert_1.default.deepStrictEqual(vars, { x: 123 });
    });
    it('should fail with non-matching repeated var', function () {
        const pattern = { a: { $var: 'x' }, b: { $var: 'x' } };
        const node = { a: 123, b: 456 };
        const vars = new PatternMatcher_1.PatternMatcher().match(node, pattern);
        assert_1.default.strictEqual(vars, null);
    });
    it('should parse staking descriptor', function () {
        const builder = new StakingDescriptorBuilder(key32('user'), [key32('provider1')], [key32('guardian1'), key32('guardian2')], 2, 100);
        const descriptor = builder.getStakingDescriptor();
        const parsed = parseStakingDescriptor(descriptor);
        assert_1.default.notStrictEqual(parsed, null);
        if (parsed) {
            assert_1.default.deepStrictEqual(parsed.penaltyNode, builder.getPenaltyMiniscriptNode());
            assert_1.default.deepStrictEqual(parsed.withdrawalNode, builder.getWithdrawalMiniscriptNode());
            assert_1.default.deepStrictEqual(parsed.timelockNode, builder.getTimelockMiniscriptNode());
        }
    });
});
