// @bitgo/sdk-coin-apt/dist/src/apt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apt = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
const utils_1 = __importDefault(require("./lib/utils"));
const _ = __importStar(require("lodash"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
class Apt extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Apt(bitgo, staticsCoin);
    }
    /**
     * Factor between the coin's base unit and its smallest subdivison
     */
    getBaseFactor() {
        return 1e8;
    }
    getChain() {
        return 'apt';
    }
    getFamily() {
        return 'apt';
    }
    getFullName() {
        return 'Aptos';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    getMPCAlgorithm() {
        return 'eddsa';
    }
    allowsAccountConsolidations() {
        return true;
    }
    async verifyTransaction(params) {
        const { txPrebuild: txPrebuild, txParams: txParams } = params;
        const txHex = txPrebuild.txHex;
        if (!txHex) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const explainedTx = await this.explainTransaction({ txHex });
        if (!explainedTx) {
            throw new Error('failed to explain transaction');
        }
        if (txParams.recipients !== undefined) {
            const filteredRecipients = txParams.recipients?.map((recipient) => {
                return {
                    address: recipient.address,
                    amount: BigInt(recipient.amount),
                };
            });
            const filteredOutputs = explainedTx.outputs.map((output) => {
                return {
                    address: output.address,
                    amount: BigInt(output.amount),
                };
            });
            if (!_.isEqual(filteredOutputs, filteredRecipients)) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            let totalAmount = new bignumber_js_1.default(0);
            for (const recipients of txParams.recipients) {
                totalAmount = totalAmount.plus(recipients.amount);
            }
            if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                throw new Error('Tx total amount does not match with expected total amount field');
            }
        }
        return true;
    }
    /**
     * Verify that an address belongs to this wallet.
     *
     * @param params - Verification parameters including address, keychains, and index
     * @returns True if address belongs to wallet
     * @throws UnexpectedAddressError if address doesn't match derived address
     */
    async isWalletAddress(params) {
        const isValid = await (0, sdk_core_1.verifyEddsaTssWalletAddress)(params, this.isValidAddress.bind(this), (pubKey) => utils_1.default.getAddressFromPublicKey(pubKey));
        if (!isValid) {
            throw new sdk_core_1.UnexpectedAddressError();
        }
        return true;
    }
    async parseTransaction(params) {
        const transactionExplanation = await this.explainTransaction({ txHex: params.txHex });
        if (!transactionExplanation) {
            throw new Error('Invalid transaction');
        }
        return {
            inputs: [
                {
                    address: transactionExplanation.sender,
                    amount: transactionExplanation.outputAmount,
                },
            ],
            outputs: [
                {
                    address: transactionExplanation.outputs[0].address,
                    amount: transactionExplanation.outputs[0].amount,
                },
            ],
        };
    }
    /**
     * Explain a Apt transaction
     * @param params
     */
    async explainTransaction(params) {
        let rebuiltTransaction;
        try {
            rebuiltTransaction = await this.rebuildTransaction(params.txHex);
        }
        catch {
            return undefined;
        }
        return rebuiltTransaction.explainTransaction();
    }
    generateKeyPair(seed) {
        const keyPair = seed ? new lib_1.KeyPair({ seed }) : new lib_1.KeyPair();
        const keys = keyPair.getKeys();
        if (!keys.prv) {
            throw new Error('Missing prv in key generation.');
        }
        return {
            pub: keys.pub,
            prv: keys.prv,
        };
    }
    isValidPub(pub) {
        return utils_1.default.isValidPublicKey(pub);
    }
    isValidAddress(address) {
        return utils_1.default.isValidAddress(address);
    }
    signTransaction(params) {
        throw new Error('Method not implemented.');
    }
    getTxBuilderFactory() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    async rebuildTransaction(txHex) {
        const txBuilderFactory = this.getTxBuilderFactory();
        try {
            const txBuilder = txBuilderFactory.from(txHex);
            return txBuilder.getRebuiltTransaction();
        }
        catch {
            throw new Error('Failed to rebuild transaction');
        }
    }
    /** @inheritDoc */
    auditDecryptedKey({ multiSigType, prv, publicKey }) {
        if (multiSigType !== 'tss') {
            throw new Error('Unsupported multiSigType');
        }
        (0, sdk_lib_mpc_1.auditEddsaPrivateKey)(prv, publicKey ?? '');
    }
    setCoinSpecificFieldsInIntent(intent, params) {
        // Handle custom transaction parameters for Aptos
        if (params.aptosCustomTransactionParams) {
            intent.aptosCustomTransactionParams = params.aptosCustomTransactionParams;
        }
    }
}
exports.Apt = Apt;
