// @bitgo/sdk-coin-apt/dist/src/aptNFTCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AptNFTCollection = void 0;
const apt_1 = require("./apt");
const statics_1 = require("@bitgo/statics");
class AptNFTCollection extends apt_1.Apt {
    constructor(bitgo, nftCollectionConfig) {
        const staticsCoin = nftCollectionConfig.network === 'Mainnet' ? statics_1.coins.get('apt') : statics_1.coins.get('tapt');
        super(bitgo, staticsCoin);
        this.nftCollectionConfig = nftCollectionConfig;
    }
    static createNFTCollectionConstructor(config) {
        return (bitgo) => new AptNFTCollection(bitgo, config);
    }
    static createNFTCollectionConstructors(nftCollectionConfigs = [
        ...statics_1.tokens.bitcoin.apt.nftCollections,
        ...statics_1.tokens.testnet.apt.nftCollections,
    ]) {
        const nftCollectionCtors = [];
        for (const config of nftCollectionConfigs) {
            const nftCollectionConstructor = AptNFTCollection.createNFTCollectionConstructor(config);
            nftCollectionCtors.push({ name: config.type, coinConstructor: nftCollectionConstructor });
        }
        return nftCollectionCtors;
    }
    get name() {
        return this.nftCollectionConfig.name;
    }
    get coin() {
        return this.nftCollectionConfig.coin;
    }
    get network() {
        return this.nftCollectionConfig.network;
    }
    get nftCollectionId() {
        return this.nftCollectionConfig.nftCollectionId;
    }
    getChain() {
        return this.nftCollectionConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Apt NFT Collection';
    }
    getBaseFactor() {
        return Math.pow(10, this.nftCollectionConfig.decimalPlaces);
    }
}
exports.AptNFTCollection = AptNFTCollection;
