// bitgo/dist/test/v2/unit/staking/goStakingWalletCommon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const nock = require("nock");
const should = require("should");
const goStakingWallet_1 = require("../../fixtures/staking/goStakingWallet");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../../../src");
const sinon = require("sinon");
const coins_1 = require("../../../../src/v2/coins");
const statics_1 = require("@bitgo/statics");
describe('Go Staking Wallet Common', function () {
    const microservicesUri = sdk_core_1.Environments['mock'].uri;
    let bitgo;
    let baseCoin;
    let enterprise;
    let stakingWallet;
    const coin = 'tsol';
    const ofcCoin = `ofc${coin}`;
    before(async function () {
        bitgo = sdk_test_1.TestBitGo.decorate(src_1.BitGo, { env: 'mock', microservicesUri });
        bitgo.initializeTestVars();
        baseCoin = bitgo.coin(ofcCoin);
        baseCoin.keychains();
        const ofcToken = statics_1.tokens.testnet.ofc.tokens.filter((token) => token.type === `ofc${coin}`)[0];
        const tokenConstructor = coins_1.OfcToken.createTokenConstructor(ofcToken);
        bitgo.register(ofcToken.type, tokenConstructor);
        enterprise = new sdk_core_1.Enterprise(bitgo, baseCoin, { id: '5cf940949449412d00f53b3d92dbcaa3', name: 'Test Enterprise' });
        const walletData = {
            id: 'walletId',
            coin: ofcCoin,
            enterprise: enterprise.id,
            keys: ['5b3424f91bf349930e340175', '5b3424f91bf349930e340174', '5b3424f91bf349930e340173'],
        };
        const wallet = new sdk_core_1.Wallet(bitgo, baseCoin, walletData);
        stakingWallet = wallet.toGoStakingWallet();
        nock(microservicesUri)
            .get(`/api/v2/${ofcCoin}/key/${stakingWallet.wallet.keyIds()[0]}`)
            .reply(200, {
            id: stakingWallet.wallet.keyIds()[0],
            pub: 'xpub661MyMwAqRbcFq65dvGMeEVb81KKDRRkWkawSVesWcyevGc5gr8V27LjNfkktaMuKtM362jhgKy2eu35RdArcmmEAoULzAvgKkJpWQPvLXM',
            source: 'user',
            encryptedPrv: bitgo.encrypt({
                input: 'xprvA41z7zogVVwxVSgdKUHDy1SKmdb533PjDz7J6N6mV6uS3ze1ai8FHa8kmHScGpWmj4WggLyQjgPie1rFSruoUihUZREPSL39UNdE3BBDu76',
                password: 'passphrase',
            }),
            coinSpecific: {},
        });
        nock(microservicesUri)
            .get(`/api/v2/${ofcCoin}/key/${stakingWallet.wallet.keyIds()[1]}`)
            .reply(200, {
            id: stakingWallet.wallet.keyIds()[1],
            pub: 'xpub661MyMwAqRbcFq65dvGMeEVb81KKDRRkWkawSVesWcyevGc5gr8V27LjNfkktaMuKtM362jhgKy2eu35RdArcmmEAoULzAvgKkJpWQPvLXM',
            source: 'user',
            encryptedPrv: bitgo.encrypt({
                input: 'xprvA41z7zogVVwxVSgdKUHDy1SKmdb533PjDz7J6N6mV6uS3ze1ai8FHa8kmHScGpWmj4WggLyQjgPie1rFSruoUihUZREPSL39UNdE3BBDu76',
                password: 'passphrase',
            }),
            coinSpecific: {},
        });
        nock(microservicesUri)
            .get(`/api/v2/${ofcCoin}/key/${stakingWallet.wallet.keyIds()[2]}`)
            .reply(200, {
            id: stakingWallet.wallet.keyIds()[2],
            pub: 'xpub661MyMwAqRbcFq65dvGMeEVb81KKDRRkWkawSVesWcyevGc5gr8V27LjNfkktaMuKtM362jhgKy2eu35RdArcmmEAoULzAvgKkJpWQPvLXM',
            source: 'user',
            encryptedPrv: bitgo.encrypt({
                input: 'xprvA41z7zogVVwxVSgdKUHDy1SKmdb533PjDz7J6N6mV6uS3ze1ai8FHa8kmHScGpWmj4WggLyQjgPie1rFSruoUihUZREPSL39UNdE3BBDu76',
                password: 'passphrase',
            }),
            coinSpecific: {},
        });
    });
    const sandbox = sinon.createSandbox();
    afterEach(function () {
        sandbox.verifyAndRestore();
    });
    describe('stake', function () {
        it('should call go-staking-service to stake', async function () {
            const preview = goStakingWallet_1.default.previewGoStakingRequest(coin);
            const msScope1 = nock(microservicesUri)
                .post(`/api/go-staking/v1/${ofcCoin}/accounts/${stakingWallet.accountId}/requests/preview`, {
                amount: '1',
                clientId: 'clientId',
                type: 'STAKE',
            })
                .reply(201, preview);
            const expected = goStakingWallet_1.default.finalizeGoStakingRequest(coin, 'STAKE');
            const msScope2 = nock(microservicesUri)
                .post(`/api/go-staking/v1/${ofcCoin}/accounts/${stakingWallet.accountId}/requests/finalize`, {
                amount: '1',
                clientId: 'clientId',
                frontTransferSendRequest: {
                    halfSigned: {
                        payload: preview.payload,
                    },
                },
                type: 'STAKE',
            })
                .reply(201, expected);
            const stakingRequest = await stakingWallet.stake({
                amount: '1',
                clientId: 'clientId',
                walletPassphrase: 'passphrase',
            });
            should.exist(stakingRequest);
            stakingRequest.should.deepEqual(expected);
            msScope1.isDone().should.be.True();
            msScope2.isDone().should.be.True();
        });
    });
    describe('unstake', function () {
        it('should call go-staking-service to unstake', async function () {
            const expected = goStakingWallet_1.default.finalizeGoStakingRequest(coin, 'UNSTAKE');
            const msScope = nock(microservicesUri)
                .post(`/api/go-staking/v1/${ofcCoin}/accounts/${stakingWallet.accountId}/requests/finalize`, {
                amount: '1',
                clientId: 'clientId',
                type: 'UNSTAKE',
            })
                .reply(201, expected);
            const stakingRequest = await stakingWallet.unstake({
                amount: '1',
                clientId: 'clientId',
            });
            should.exist(stakingRequest);
            stakingRequest.should.deepEqual(expected);
            msScope.isDone().should.be.True();
        });
    });
    describe('getGoStakingRequest', function () {
        it('should call go-staking-service to get go staking request', async function () {
            const stakingRequestId = '8638284a-dab2-46b9-b07f-21109a6e7220';
            const expected = goStakingWallet_1.default.finalizeGoStakingRequest(coin, 'STAKE');
            const msScope = nock(microservicesUri)
                .get(`/api/go-staking/v1/${ofcCoin}/accounts/${stakingWallet.accountId}/requests/${stakingRequestId}`)
                .reply(200, expected);
            const stakingRequest = await stakingWallet.getGoStakingRequest(stakingRequestId);
            should.exist(stakingRequest);
            stakingRequest.should.deepEqual(expected);
            msScope.isDone().should.be.True();
        });
    });
    describe('getGoStakingRequestsByCoin', function () {
        it('should call go-staking-service to get go staking requests by coin', async function () {
            const expected = goStakingWallet_1.default.getGoStakingRequestsByCoin(ofcCoin);
            const msScope = nock(microservicesUri)
                .get(`/api/go-staking/v1/${ofcCoin}/accounts/${stakingWallet.accountId}/requests`)
                .reply(200, expected);
            const stakingRequests = await stakingWallet.getGoStakingRequestsByWalletCoin();
            should.exist(stakingRequests);
            stakingRequests.should.deepEqual(expected);
            msScope.isDone().should.be.True();
        });
    });
    describe('getGoStakingRequests', function () {
        it('should call go-staking-service to get go staking requests', async function () {
            const expected = goStakingWallet_1.default.getGoStakingRequests([ofcCoin]);
            const msScope = nock(microservicesUri)
                .get(`/api/go-staking/v1/accounts/${stakingWallet.accountId}/requests`)
                .reply(200, expected);
            const stakingRequests = await stakingWallet.getGoStakingRequests();
            should.exist(stakingRequests);
            stakingRequests.should.deepEqual(expected);
            msScope.isDone().should.be.True();
        });
    });
    describe('getGoStakingWallets', function () {
        it('should call go-staking-service to get go staking wallets', async function () {
            const expected = goStakingWallet_1.default.getGoStakingWallets([ofcCoin]);
            const msScope = nock(microservicesUri)
                .get(`/api/go-staking/v1/accounts/${stakingWallet.accountId}/coins`)
                .reply(200, expected);
            const goStakingWallets = await stakingWallet.getGoStakingWallets();
            should.exist(goStakingWallets);
            goStakingWallets.should.deepEqual(expected);
            msScope.isDone().should.be.True();
        });
    });
    describe('getGoStakingWallet', function () {
        it('should call go-staking-service to get go staking wallet', async function () {
            const expected = goStakingWallet_1.default.getGoStakingWallet(ofcCoin);
            const msScope = nock(microservicesUri)
                .get(`/api/go-staking/v1/${ofcCoin}/accounts/${stakingWallet.accountId}/coins`)
                .reply(200, expected);
            const goStakingWallet = await stakingWallet.getGoStakingWallet();
            should.exist(goStakingWallet);
            goStakingWallet.should.deepEqual(expected);
            msScope.isDone().should.be.True();
        });
    });
});
