// bitgo/dist/test/v2/unit/staking/stakingWalletTSS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const nock = require("nock");
const stakingWallet_1 = require("../../fixtures/staking/stakingWallet");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../../../src");
const sinon = require("sinon");
describe('TSS Staking Wallet', function () {
    const microservicesUri = sdk_core_1.Environments['mock'].uri;
    let bitgo;
    let nearBaseCoin;
    let ethBaseCoin;
    let atomBaseCoin;
    let enterprise;
    let ethWalletData;
    let nearStakingWallet;
    let ethStakingWallet;
    let atomStakingWallet;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(src_1.BitGo, { env: 'mock', microservicesUri });
        bitgo.initializeTestVars();
        nearBaseCoin = bitgo.coin('near');
        nearBaseCoin.keychains();
        ethBaseCoin = bitgo.coin('eth');
        ethBaseCoin.keychains();
        atomBaseCoin = bitgo.coin('atom');
        atomBaseCoin.keychains();
        enterprise = new sdk_core_1.Enterprise(bitgo, nearBaseCoin, {
            id: '5cf940949449412d00f53b3d92dbcaa3',
            name: 'TSS Test Enterprise',
        });
        const tssWalletData = {
            id: 'walletIdTss',
            coin: 'near',
            enterprise: enterprise.id,
            keys: ['5b3424f91bf349930e340175'],
            multisigType: 'tss',
        };
        const nearWallet = new sdk_core_1.Wallet(bitgo, nearBaseCoin, tssWalletData);
        nearStakingWallet = nearWallet.toStakingWallet();
        ethWalletData = {
            id: 'walletId',
            coin: 'eth',
            enterprise: enterprise.id,
            keys: ['5b3424f91bf349930e340175'],
            coinSpecific: { walletVersion: 3 },
        };
        const ethWallet = new sdk_core_1.Wallet(bitgo, ethBaseCoin, ethWalletData);
        ethStakingWallet = ethWallet.toStakingWallet();
        const atomWallet = new sdk_core_1.Wallet(bitgo, atomBaseCoin, { ...tssWalletData, coin: 'atom' });
        atomStakingWallet = atomWallet.toStakingWallet();
    });
    describe('buildSignAndSend', function () {
        const sandbox = sinon.createSandbox();
        afterEach(function () {
            sandbox.verifyAndRestore();
        });
        it('should throw error when txRequestId is not defined', async function () {
            const transaction = stakingWallet_1.default.transaction('READY');
            transaction.txRequestId = undefined;
            await nearStakingWallet
                .buildSignAndSend({ walletPassphrase: 'passphrase' }, transaction)
                .should.rejectedWith('txRequestId is required to sign and send');
        });
        it('should build, sign and send transaction', async function () {
            const walletPassphrase = 'passphrase';
            const transaction = stakingWallet_1.default.transaction('READY');
            const deleteSignatureShares = sandbox.stub(sdk_core_1.TssUtils.prototype, 'deleteSignatureShares');
            deleteSignatureShares.resolves([]);
            deleteSignatureShares.calledOnceWithExactly(transaction.id);
            const getKeysForSigning = sandbox.stub(sdk_core_1.Keychains.prototype, 'getKeysForSigning');
            const keyChain = {
                id: 'id',
                pub: 'pub',
                type: 'tss',
            };
            getKeysForSigning.resolves([keyChain]);
            getKeysForSigning.calledOnce;
            const signTransaction = sandbox.stub(sdk_core_1.Wallet.prototype, 'signTransaction');
            signTransaction.resolves({ txRequestId: stakingWallet_1.default.txRequestId });
            signTransaction.calledOnceWithExactly({
                txPrebuild: {
                    txRequestId: stakingWallet_1.default.txRequestId,
                },
                walletPassphrase: walletPassphrase,
                keychain: keyChain,
            });
            nock(microservicesUri)
                .post(`/api/staking/v1/${nearStakingWallet.coin}/wallets/${nearStakingWallet.walletId}/requests/${transaction.stakingRequestId}/transactions/${transaction.id}`, _.matches({ txRequestId: stakingWallet_1.default.txRequestId }))
                .reply(200, transaction);
            const stakingTransaction = await nearStakingWallet.buildSignAndSend({ walletPassphrase: walletPassphrase }, transaction);
            stakingTransaction.should.deepEqual(transaction);
        });
        it('should build and sign but not send transaction for ETH TSS or ECDSA based TSS Coin', async function () {
            [ethStakingWallet, atomStakingWallet].forEach(async (ecdsaStakingWallet) => {
                const walletPassphrase = 'passphrase';
                const transaction = stakingWallet_1.default.transaction('READY');
                const deleteSignatureShares = sandbox.stub(sdk_core_1.TssUtils.prototype, 'deleteSignatureShares');
                deleteSignatureShares.resolves([]);
                deleteSignatureShares.calledOnceWithExactly(transaction.id);
                const getKeysForSigning = sandbox.stub(sdk_core_1.Keychains.prototype, 'getKeysForSigning');
                const keyChain = {
                    id: 'id',
                    pub: 'pub',
                    type: 'tss',
                };
                getKeysForSigning.resolves([keyChain]);
                getKeysForSigning.calledOnce;
                const signTransaction = sandbox.stub(sdk_core_1.Wallet.prototype, 'signTransaction');
                signTransaction.resolves({ txRequestId: stakingWallet_1.default.txRequestId });
                signTransaction.calledOnceWithExactly({
                    txPrebuild: {
                        txRequestId: stakingWallet_1.default.txRequestId,
                    },
                    walletPassphrase: walletPassphrase,
                    keychain: keyChain,
                });
                const stakingTransaction = await ecdsaStakingWallet.buildSignAndSend({ walletPassphrase: walletPassphrase }, transaction);
                stakingTransaction.should.deepEqual(transaction);
            });
        });
    });
});
