// @bitgo/sdk-core/dist/test/unit/bitgo/errors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const errors_1 = require("../../../src/bitgo/errors");
describe('Transaction Intent Mismatch Errors', () => {
    const mockTransactionId = '0x1234567890abcdef';
    const mockTxParams = [
        { recipients: [{ address: '0xrecipient1', amount: '1000000000000000000' }] },
        { recipients: [{ address: '0xrecipient2', amount: '2000000000000000000' }] },
    ];
    const mockTxHex = '0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef';
    describe('TxIntentMismatchError', () => {
        it('should create base transaction intent mismatch error with all required properties', () => {
            const message = 'Transaction does not match user intent';
            const error = new errors_1.TxIntentMismatchError(message, mockTransactionId, mockTxParams, mockTxHex);
            should_1.default.exist(error);
            should_1.default.equal(error.message, message);
            should_1.default.equal(error.name, 'TxIntentMismatchError');
            should_1.default.equal(error.id, mockTransactionId);
            should_1.default.deepEqual(error.txParams, mockTxParams);
            should_1.default.equal(error.txHex, mockTxHex);
            should_1.default.equal(error.txExplanation, undefined); // txExplanation is optional
        });
        it('should create error with txExplanation when provided', () => {
            const message = 'Transaction does not match user intent';
            const txExplanationObj = {
                id: '0xtxid',
                outputAmount: '1000000',
                outputs: [{ address: '0xrecipient', amount: '1000000' }],
                fee: { fee: '21000' },
            };
            const error = new errors_1.TxIntentMismatchError(message, mockTransactionId, mockTxParams, mockTxHex, txExplanationObj);
            should_1.default.exist(error);
            // Should be stringified
            should_1.default.equal(typeof error.txExplanation, 'string');
            const parsed = JSON.parse(error.txExplanation);
            should_1.default.equal(parsed.id, '0xtxid');
            should_1.default.equal(parsed.outputAmount, '1000000');
        });
        it('should be an instance of Error', () => {
            const error = new errors_1.TxIntentMismatchError('Test message', mockTransactionId, mockTxParams, mockTxHex);
            (0, should_1.default)(error).be.instanceOf(Error);
        });
    });
    describe('TxIntentMismatchRecipientError', () => {
        it('should create recipient intent mismatch error with mismatched recipients', () => {
            const message = 'Transaction recipients do not match user intent';
            const mismatchedRecipients = [
                { address: '0xexpected1', amount: '1000' },
                { address: '0xexpected2', amount: '2000' },
            ];
            const error = new errors_1.TxIntentMismatchRecipientError(message, mockTransactionId, mockTxParams, mockTxHex, mismatchedRecipients);
            should_1.default.exist(error);
            should_1.default.equal(error.message, message);
            should_1.default.equal(error.name, 'TxIntentMismatchRecipientError');
            should_1.default.equal(error.id, mockTransactionId);
            should_1.default.deepEqual(error.txParams, mockTxParams);
            should_1.default.equal(error.txHex, mockTxHex);
            should_1.default.deepEqual(error.mismatchedRecipients, mismatchedRecipients);
            should_1.default.equal(error.txExplanation, undefined); // txExplanation is optional
        });
        it('should create recipient error with txExplanation when provided', () => {
            const message = 'Transaction recipients do not match user intent';
            const mismatchedRecipients = [{ address: '0xexpected1', amount: '1000' }];
            const txExplanationObj = {
                id: '0xtxid',
                outputs: [{ address: '0xactual', amount: '1000' }],
            };
            const error = new errors_1.TxIntentMismatchRecipientError(message, mockTransactionId, mockTxParams, mockTxHex, mismatchedRecipients, txExplanationObj);
            should_1.default.equal(typeof error.txExplanation, 'string');
            const parsed = JSON.parse(error.txExplanation);
            should_1.default.equal(parsed.id, '0xtxid');
            should_1.default.deepEqual(error.mismatchedRecipients, mismatchedRecipients);
        });
        it('should be an instance of TxIntentMismatchError', () => {
            const error = new errors_1.TxIntentMismatchRecipientError('Test message', mockTransactionId, mockTxParams, mockTxHex, []);
            (0, should_1.default)(error).be.instanceOf(errors_1.TxIntentMismatchError);
            (0, should_1.default)(error).be.instanceOf(Error);
        });
    });
    describe('TxIntentMismatchContractError', () => {
        it('should create contract intent mismatch error with mismatched data payload', () => {
            const message = 'Contract interaction does not match user intent';
            const mismatchedDataPayload = {
                address: '0xcontract123',
                rawContractPayload: '0xabcdef',
                decodedContractPayload: { method: 'transfer', params: ['0xrecipient', '1000'] },
            };
            const error = new errors_1.TxIntentMismatchContractError(message, mockTransactionId, mockTxParams, mockTxHex, mismatchedDataPayload);
            should_1.default.exist(error);
            should_1.default.equal(error.message, message);
            should_1.default.equal(error.name, 'TxIntentMismatchContractError');
            should_1.default.equal(error.id, mockTransactionId);
            should_1.default.deepEqual(error.txParams, mockTxParams);
            should_1.default.equal(error.txHex, mockTxHex);
            should_1.default.deepEqual(error.mismatchedDataPayload, mismatchedDataPayload);
            should_1.default.equal(error.txExplanation, undefined); // txExplanation is optional
        });
        it('should create contract error with txExplanation when provided', () => {
            const message = 'Contract interaction does not match user intent';
            const mismatchedDataPayload = {
                address: '0xcontract123',
                rawContractPayload: '0xabcdef',
                decodedContractPayload: { method: 'approve', params: ['0xspender', 'unlimited'] },
            };
            const txExplanationObj = {
                id: '0xtxid',
                outputs: [{ address: '0xcontract123', amount: '0' }],
                contractCall: { method: 'approve', params: ['0xspender', 'unlimited'] },
            };
            const error = new errors_1.TxIntentMismatchContractError(message, mockTransactionId, mockTxParams, mockTxHex, mismatchedDataPayload, txExplanationObj);
            should_1.default.equal(typeof error.txExplanation, 'string');
            const parsed = JSON.parse(error.txExplanation);
            should_1.default.equal(parsed.id, '0xtxid');
            should_1.default.deepEqual(error.mismatchedDataPayload, mismatchedDataPayload);
        });
        it('should be an instance of TxIntentMismatchError', () => {
            const error = new errors_1.TxIntentMismatchContractError('Test message', mockTransactionId, mockTxParams, mockTxHex, {
                address: '0xtest',
                rawContractPayload: '0x',
                decodedContractPayload: {},
            });
            (0, should_1.default)(error).be.instanceOf(errors_1.TxIntentMismatchError);
            (0, should_1.default)(error).be.instanceOf(Error);
        });
    });
    describe('TxIntentMismatchApprovalError', () => {
        it('should create approval intent mismatch error with token approval details', () => {
            const message = 'Token approval does not match user intent';
            const tokenApproval = {
                tokenName: 'TestToken',
                tokenAddress: '0xtoken123',
                authorizingAmount: { type: 'unlimited' },
                authorizingAddress: '0xspender456',
            };
            const error = new errors_1.TxIntentMismatchApprovalError(message, mockTransactionId, mockTxParams, mockTxHex, tokenApproval);
            should_1.default.exist(error);
            should_1.default.equal(error.message, message);
            should_1.default.equal(error.name, 'TxIntentMismatchApprovalError');
            should_1.default.equal(error.id, mockTransactionId);
            should_1.default.deepEqual(error.txParams, mockTxParams);
            should_1.default.equal(error.txHex, mockTxHex);
            should_1.default.deepEqual(error.tokenApproval, tokenApproval);
            should_1.default.equal(error.txExplanation, undefined); // txExplanation is optional
        });
        it('should create approval error with txExplanation when provided', () => {
            const message = 'Token approval does not match user intent';
            const tokenApproval = {
                tokenName: 'USDC',
                tokenAddress: '0xusdc',
                authorizingAmount: { type: 'unlimited' },
                authorizingAddress: '0xmalicious',
            };
            const txExplanationObj = {
                id: '0xtxid',
                outputs: [{ address: '0xusdc', amount: '0' }],
                tokenApproval: {
                    token: 'USDC',
                    spender: '0xmalicious',
                    amount: 'unlimited',
                },
            };
            const error = new errors_1.TxIntentMismatchApprovalError(message, mockTransactionId, mockTxParams, mockTxHex, tokenApproval, txExplanationObj);
            should_1.default.equal(typeof error.txExplanation, 'string');
            const parsed = JSON.parse(error.txExplanation);
            should_1.default.equal(parsed.id, '0xtxid');
            should_1.default.deepEqual(error.tokenApproval, tokenApproval);
        });
        it('should be an instance of TxIntentMismatchError', () => {
            const error = new errors_1.TxIntentMismatchApprovalError('Test message', mockTransactionId, mockTxParams, mockTxHex, {
                tokenAddress: '0xtoken',
                authorizingAmount: { type: 'limited', amount: 1000 },
                authorizingAddress: '0xspender',
            });
            (0, should_1.default)(error).be.instanceOf(errors_1.TxIntentMismatchError);
            (0, should_1.default)(error).be.instanceOf(Error);
        });
    });
    describe('Error inheritance and properties', () => {
        it('should maintain proper inheritance chain', () => {
            const baseError = new errors_1.TxIntentMismatchError('Base error', mockTransactionId, mockTxParams, mockTxHex);
            const recipientError = new errors_1.TxIntentMismatchRecipientError('Recipient error', mockTransactionId, mockTxParams, mockTxHex, []);
            const contractError = new errors_1.TxIntentMismatchContractError('Contract error', mockTransactionId, mockTxParams, mockTxHex, { address: '0xtest', rawContractPayload: '0x', decodedContractPayload: {} });
            const approvalError = new errors_1.TxIntentMismatchApprovalError('Approval error', mockTransactionId, mockTxParams, mockTxHex, {
                tokenAddress: '0xtoken',
                authorizingAmount: { type: 'limited', amount: 1000 },
                authorizingAddress: '0xspender',
            });
            // All should be instances of Error
            (0, should_1.default)(baseError).be.instanceOf(Error);
            (0, should_1.default)(recipientError).be.instanceOf(Error);
            (0, should_1.default)(contractError).be.instanceOf(Error);
            (0, should_1.default)(approvalError).be.instanceOf(Error);
            // All should be instances of TxIntentMismatchError
            (0, should_1.default)(recipientError).be.instanceOf(errors_1.TxIntentMismatchError);
            (0, should_1.default)(contractError).be.instanceOf(errors_1.TxIntentMismatchError);
            (0, should_1.default)(approvalError).be.instanceOf(errors_1.TxIntentMismatchError);
        });
        it('should preserve stack trace', () => {
            const error = new errors_1.TxIntentMismatchError('Test error', mockTransactionId, mockTxParams, mockTxHex);
            should_1.default.exist(error.stack);
            (0, should_1.default)(error.stack).be.a.String();
            (0, should_1.default)(error.stack).containEql('TxIntentMismatchError');
        });
    });
    describe('Transaction explanation property', () => {
        it('should handle valid JSON transaction explanations', () => {
            const txExplanationObj = {
                id: '0x123abc',
                outputAmount: '1000000000000000000',
                outputs: [
                    { address: '0xrecipient1', amount: '500000000000000000' },
                    { address: '0xrecipient2', amount: '500000000000000000' },
                ],
                fee: { fee: '21000', gasLimit: '21000' },
                type: 'send',
            };
            const error = new errors_1.TxIntentMismatchError('Test', mockTransactionId, mockTxParams, mockTxHex, txExplanationObj);
            should_1.default.equal(typeof error.txExplanation, 'string');
            // Verify it can be parsed back
            const parsed = JSON.parse(error.txExplanation);
            should_1.default.equal(parsed.id, '0x123abc');
            should_1.default.equal(parsed.outputs.length, 2);
        });
        it('should handle empty transaction explanation', () => {
            const error = new errors_1.TxIntentMismatchError('Test', mockTransactionId, mockTxParams, mockTxHex, undefined);
            should_1.default.equal(error.txExplanation, undefined);
        });
        it('should handle BigInt values in transaction explanation', () => {
            const txExplanationObj = {
                id: '0x123',
                outputAmount: BigInt('9007199254740992'),
                changeAmount: BigInt('1234567890123456789'),
                outputs: [{ address: 'addr1', amount: BigInt('9007199254740992') }],
            };
            const error = new errors_1.TxIntentMismatchError('Test', mockTransactionId, mockTxParams, mockTxHex, txExplanationObj);
            should_1.default.equal(typeof error.txExplanation, 'string');
            const parsed = JSON.parse(error.txExplanation);
            // BigInts should be converted to strings
            should_1.default.equal(parsed.outputAmount, '9007199254740992');
            should_1.default.equal(parsed.changeAmount, '1234567890123456789');
            should_1.default.equal(parsed.outputs[0].amount, '9007199254740992');
        });
        it('should work with all error subclasses', () => {
            const txExplanationObj = { id: '0x123', outputs: [] };
            const errors = [
                new errors_1.TxIntentMismatchRecipientError('Test', mockTransactionId, mockTxParams, mockTxHex, [], txExplanationObj),
                new errors_1.TxIntentMismatchContractError('Test', mockTransactionId, mockTxParams, mockTxHex, { address: '0x', rawContractPayload: '0x', decodedContractPayload: {} }, txExplanationObj),
                new errors_1.TxIntentMismatchApprovalError('Test', mockTransactionId, mockTxParams, mockTxHex, { tokenAddress: '0x', authorizingAmount: { type: 'limited', amount: 0 }, authorizingAddress: '0x' }, txExplanationObj),
            ];
            errors.forEach((error) => {
                should_1.default.equal(typeof error.txExplanation, 'string');
                const parsed = JSON.parse(error.txExplanation);
                should_1.default.equal(parsed.id, '0x123');
            });
        });
    });
});
