// @bitgo/abstract-utxo/dist/cjs/test/unit/util/keychains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keychains = exports.keychainsBase58 = exports.KeyNames = void 0;
exports.toKeychainBase58 = toKeychainBase58;
exports.toKeychainObjects = toKeychainObjects;
exports.getWalletUnspentSignerUserBitGo = getWalletUnspentSignerUserBitGo;
exports.getDefaultWalletKeys = getDefaultWalletKeys;
exports.getDefaultWalletUnspentSigner = getDefaultWalletUnspentSigner;
exports.encryptKeychain = encryptKeychain;
exports.getWalletKeys = getWalletKeys;
exports.createTestWalletKeys = createTestWalletKeys;
exports.createWasmWalletKeys = createWasmWalletKeys;
exports.getDefaultWasmWalletKeys = getDefaultWasmWalletKeys;
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const utxo_lib_1 = require("@bitgo/utxo-lib");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
function toKeychainBase58(k) {
    if (k.isNeutered()) {
        throw new Error(`must provide private key`);
    }
    return {
        prv: k.toBase58(),
        pub: k.neutered().toBase58(),
    };
}
function toKeychainObjects(rootWalletKeys, walletPassphrase) {
    return rootWalletKeys.triple.map((bip32, keyIdx) => {
        const pub = bip32.neutered().toBase58();
        return {
            id: (0, sdk_test_1.getSeed)(pub).toString('hex'),
            pub,
            source: exports.KeyNames[keyIdx],
            encryptedPrv: (0, sdk_api_1.encrypt)(walletPassphrase, bip32.toBase58()),
            coinSpecific: {},
        };
    });
}
exports.KeyNames = ['user', 'backup', 'bitgo'];
exports.keychainsBase58 = [
    {
        pub: 'xpub661MyMwAqRbcGiQhVk1J7cD1YodF9tc5Y1B8vpTjjB1pcB1J1m1QX8fMtYP2sYqFmW6J2ra69tNoARKjvTGo9cGUrbPbJdjwrSzGGzPzWWS',
        prv: 'xprv9s21ZrQH143K4ELEPiUHkUGGzmnkkRtEAnFY8S48AqUqjNg9UDh9yLLt3FcfATyCjbsMB9JCGHAD8MeBTAK1P7trFppkoswu5ZAsHYASfbk',
    },
    {
        pub: 'xpub661MyMwAqRbcFzLXuganogQvd7MrefQQqCcJP2ZDumnCdQecf5cw1P1nD5qBz8SNS1yCLSC9VqpNUWnQU3V6qmnPt2r21oXhicQFzPA6Lby',
        prv: 'xprv9s21ZrQH143K3WG4of3nSYUC55XNFCgZTyghae9cMSFDkcKU7YJgTahJMpdTY9CjCcjgSo2TJ635uUVx176BufUMBFpieKYVJD9J3VvrGRm',
    },
    {
        pub: 'xpub661MyMwAqRbcFHpwWrzPB61U2CgBmdD21WNVM1JKUn9rEExkoGE4yafUVFbPSd78vdX8tWcEUQWaALFkU9fUbUM4Cc49DKEJSCYGRnbzCym',
        prv: 'xprv9s21ZrQH143K2okUQqTNox4jUAqhNAVAeHStYcthvScsMSdcFiupRnLzdxzfJithak5Zs92FQJeeJ9Jiya63KfUNxawuMZDCp2cGT9cdMKs',
    },
];
exports.keychains = exports.keychainsBase58.map(({ pub, prv }) => {
    const k = utxo_lib_1.bip32.fromBase58(prv);
    if (k.neutered().toBase58() !== pub) {
        throw new Error(`mismatch`);
    }
    return k;
});
function getWalletUnspentSignerUserBitGo(keys) {
    return new utxo_lib_1.bitgo.WalletUnspentSigner(keys, keys[0], keys[2]);
}
function getDefaultWalletKeys() {
    return new utxo_lib_1.bitgo.RootWalletKeys(exports.keychains);
}
function getDefaultWalletUnspentSigner() {
    return getWalletUnspentSignerUserBitGo(exports.keychains);
}
function encryptKeychain(password, keychain) {
    return (0, sdk_api_1.encrypt)(password, keychain.prv);
}
function getWalletKeys(seed) {
    return new utxo_lib_1.bitgo.RootWalletKeys(Array.from({ length: 3 }).map((_, i) => utxo_lib_1.bip32.fromSeed((0, sdk_test_1.getSeed)(`${seed}/${i}`))));
}
/**
 * Create test wallet keys from a seed string.
 * Uses the same seed generation as getWalletKeys (getSeed('seed/i'))
 * to ensure compatibility with existing test fixtures.
 */
function createTestWalletKeys(seed) {
    const keys = Array.from({ length: 3 }).map((_, i) => utxo_lib_1.bip32.fromSeed((0, sdk_test_1.getSeed)(`${seed}/${i}`)));
    return {
        xpubs: keys.map((k) => k.neutered().toBase58()),
        xprivs: keys.map((k) => k.toBase58()),
    };
}
/**
 * Create wasm-utxo RootWalletKeys from a seed string.
 * Optionally specify custom derivation prefixes.
 */
function createWasmWalletKeys(seed, derivationPrefixes = ['m/0/0', 'm/0/0', 'm/0/0']) {
    const { xpubs: xpubStrs, xprivs: xprivStrs } = createTestWalletKeys(seed);
    const xpubs = xpubStrs.map((xpub) => wasm_utxo_1.BIP32.from(xpub));
    const xprivs = xprivStrs.map((xpriv) => wasm_utxo_1.BIP32.from(xpriv));
    return {
        walletKeys: wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from({
            triple: xpubs,
            derivationPrefixes,
        }),
        xpubs,
        xprivs,
    };
}
/**
 * Get wasm-utxo RootWalletKeys from the default hardcoded keychains.
 * Use this when you need wallet keys that match existing test fixtures.
 * Optionally specify custom derivation prefixes.
 */
function getDefaultWasmWalletKeys(derivationPrefixes = ['m/0/0', 'm/0/0', 'm/0/0']) {
    const xpubs = exports.keychainsBase58.map((k) => wasm_utxo_1.BIP32.from(k.pub));
    const xprivs = exports.keychainsBase58.map((k) => wasm_utxo_1.BIP32.from(k.prv));
    return {
        walletKeys: wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from({
            triple: xpubs,
            derivationPrefixes,
        }),
        xpubs,
        xprivs,
    };
}
