// @bitgo/abstract-utxo/dist/cjs/test/unit/util/nockIndexerAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nockBitGoPublicTransaction = nockBitGoPublicTransaction;
exports.nockBitGoPublicAddressUnspents = nockBitGoPublicAddressUnspents;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const nockBitGo_1 = require("./nockBitGo");
function nockBitGoPublicTransaction(coin, tx, unspents) {
    const payload = {
        input: unspents.map((u) => ({ address: u.address })),
        outputs: tx.outs.map((o) => ({ address: wasm_utxo_1.address.fromOutputScriptWithCoin(o.script, coin.name) })),
    };
    return (0, nockBitGo_1.nockBitGo)().get(`/api/v2/${coin.getChain()}/public/tx/${tx.getId()}`).reply(200, payload);
}
function nockBitGoPublicAddressUnspents(coin, txid, address, outputs) {
    const payload = outputs.map((o, vout) => ({
        id: `${txid}:${vout}`,
        address: wasm_utxo_1.address.fromOutputScriptWithCoin(o.script, coin.name),
        value: Number(o.value),
        valueString: coin.amountType === 'bigint' ? o.value.toString() : undefined,
    }));
    return (0, nockBitGo_1.nockBitGo)().get(`/api/v2/${coin.getChain()}/public/addressUnspents/${address}`).reply(200, payload);
}
