// @bitgo/abstract-utxo/dist/cjs/test/unit/util/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertEqualParsedPsbt = assertEqualParsedPsbt;
exports.transactionToObj = transactionToObj;
exports.transactionHexToObj = transactionHexToObj;
exports.createPrebuildTransaction = createPrebuildTransaction;
exports.createHalfSignedTransaction = createHalfSignedTransaction;
exports.createFullSignedTransaction = createFullSignedTransaction;
const assert_1 = __importDefault(require("assert"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const utxoCoins_1 = require("./utxoCoins");
const { isWalletUnspent, signInputWithUnspent } = utxolib.bitgo;
function toTxOutput(u, network) {
    return {
        script: Buffer.from(wasm_utxo_1.address.toOutputScriptWithCoin(u.address, (0, utxoCoins_1.getCoinNameForNetwork)(network))),
        value: u.value,
    };
}
function assertEqualParsedPsbt(a, b, coinName, walletKeys, replayProtection) {
    if (!(0, wasm_utxo_1.hasPsbtMagic)(a)) {
        throw new Error('a is not a psbt');
    }
    if (!(0, wasm_utxo_1.hasPsbtMagic)(b)) {
        throw new Error('b is not a psbt');
    }
    const aPsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(a, coinName);
    const bPsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(b, coinName);
    const aParsed = aPsbt.parseTransactionWithWalletKeys(walletKeys, {
        replayProtection: { publicKeys: replayProtection },
    });
    const bParsed = bPsbt.parseTransactionWithWalletKeys(walletKeys, {
        replayProtection: { publicKeys: replayProtection },
    });
    assert_1.default.deepStrictEqual(aParsed, bParsed);
}
function transactionToObj(tx) {
    return {
        id: tx.getId(),
        hex: tx.toBuffer().toString('hex'),
        ins: tx.ins.map((v) => ({
            txid: Buffer.from(v.hash).reverse().toString('hex'),
            index: v.index,
            script: v.script?.toString('hex'),
            witness: v.witness?.map((v) => v.toString('hex')),
        })),
        outs: tx.outs.map((v) => ({
            script: v.script.toString('hex'),
            value: typeof v.value === 'bigint' ? v.value.toString() : v.value,
        })),
    };
}
function transactionHexToObj(txHex, network, amountType = 'number') {
    const obj = transactionToObj(utxolib.bitgo.createTransactionFromBuffer(Buffer.from(txHex, 'hex'), network, { amountType }));
    if (obj.hex !== txHex) {
        throw new Error(`serialized txHex does not match input`);
    }
    return obj;
}
function createPrebuildTransaction(network, unspents, outputAddress) {
    const txb = utxolib.bitgo.createTransactionBuilderForNetwork(network);
    unspents.forEach((u) => {
        const [txid, vin] = u.id.split(':');
        txb.addInput(txid, Number(vin));
    });
    const amountType = unspents.length > 0 && typeof unspents[0].value === 'bigint' ? 'bigint' : 'number';
    if (amountType === 'number') {
        unspents.forEach((u) => (u.value = Math.round(u.value)));
    }
    const unspentSum = utxolib.bitgo.unspentSum(unspents, amountType);
    txb.addOutput(outputAddress, utxolib.bitgo.toTNumber(BigInt(unspentSum) - BigInt(1000), amountType));
    return txb.buildIncomplete();
}
function createTransactionBuilderWithSignedInputs(network, unspents, signer, inputTransaction) {
    const txBuilder = utxolib.bitgo.createTransactionBuilderFromTransaction(inputTransaction, unspents.map((u) => toTxOutput(u, network)));
    unspents.forEach((u, inputIndex) => {
        if (isWalletUnspent(u)) {
            signInputWithUnspent(txBuilder, inputIndex, u, signer);
        }
    });
    return txBuilder;
}
function createHalfSignedTransaction(network, unspents, outputAddress, signer, prebuild) {
    if (!prebuild) {
        prebuild = createPrebuildTransaction(network, unspents, outputAddress);
    }
    return createTransactionBuilderWithSignedInputs(network, unspents, signer, prebuild).buildIncomplete();
}
function createFullSignedTransaction(network, unspents, outputAddress, signer, halfSigned) {
    if (!halfSigned) {
        halfSigned = createHalfSignedTransaction(network, unspents, outputAddress, signer);
    }
    return createTransactionBuilderWithSignedInputs(network, unspents, new utxolib.bitgo.WalletUnspentSigner(signer.walletKeys, signer.cosigner, signer.signer), halfSigned).build();
}
