// @bitgo/abstract-utxo/dist/cjs/test/unit/util/unspents.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockWalletUnspent = mockWalletUnspent;
exports.mockUnspentReplayProtection = mockUnspentReplayProtection;
exports.mockUnspent = mockUnspent;
exports.toUnspent = toUnspent;
exports.createWasmWalletUnspent = createWasmWalletUnspent;
exports.createMockPrevTx = createMockPrevTx;
exports.toUnspentWithPrevTx = toUnspentWithPrevTx;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const sdk_test_1 = require("@bitgo/sdk-test");
const wasmUtxo = __importStar(require("@bitgo/wasm-utxo"));
const src_1 = require("../../../src");
const names_1 = require("../../../src/names");
const utxoCoins_1 = require("./utxoCoins");
const address_1 = require("./address");
const { chainCodesP2sh, getExternalChainCode, getInternalChainCode } = utxolib.bitgo;
function toCoinName(network) {
    if (typeof network === 'string') {
        if (!(0, names_1.isUtxoCoinName)(network)) {
            throw new Error(`not a valid coin name: ${network}`);
        }
        return network;
    }
    return (0, utxoCoins_1.getCoinNameForNetwork)(network);
}
const defaultChain = getExternalChainCode(chainCodesP2sh);
function mockOutputIdForAddress(address) {
    return (0, sdk_test_1.getSeed)(address).toString('hex') + ':1';
}
function mockWalletUnspent(network, walletKeys, { id, chain = defaultChain, index = 0, value, address }) {
    if (value === undefined) {
        throw new Error(`unspent value must be set`);
    }
    if (chain === undefined) {
        throw new Error(`unspent chain must be set`);
    }
    const deriveAddress = (0, address_1.getWalletAddress)(toCoinName(network), walletKeys, chain, index);
    if (address) {
        if (address !== deriveAddress) {
            throw new Error(`derivedAddress mismatch: ${address} derived=${deriveAddress}`);
        }
    }
    else {
        address = deriveAddress;
    }
    if (!id) {
        id = mockOutputIdForAddress(address);
    }
    return {
        id,
        address,
        chain,
        index,
        value,
    };
}
function mockUnspentReplayProtection(network, value) {
    const coinName = toCoinName(network);
    const addresses = (0, src_1.getReplayProtectionAddresses)(coinName);
    if (addresses.length) {
        const address = addresses[0];
        return {
            id: mockOutputIdForAddress(address),
            address,
            value,
        };
    }
    throw new Error(`${coinName} has no replay protection unspents`);
}
function mockUnspent(network, walletKeys, chain, index, value) {
    if (chain === 'replayProtection') {
        return mockUnspentReplayProtection(network, (typeof value === 'bigint' ? BigInt(1000) : 1000));
    }
    else {
        return mockWalletUnspent(network, walletKeys, { chain: getInternalChainCode(chain), value, index });
    }
}
/**
 * Create unspent object from input script type, index, network and root wallet keys.
 * Supports both utxolib and wasm-utxo RootWalletKeys.
 */
function toUnspent(input, index, network, rootWalletKeys) {
    const scriptType = input.scriptType === 'taprootKeyPathSpend' ? 'p2trMusig2' : input.scriptType;
    const chain = getInternalChainCode(scriptType);
    return mockWalletUnspent(network, rootWalletKeys, {
        chain,
        index,
        value: input.value,
    });
}
/**
 * Create a mock wallet unspent using wasm-utxo only (no utxolib dependency).
 * Uses createMockPrevTx to generate a realistic txid for the unspent.
 *
 * This is useful for tests that need a valid unspent but don't require the prevTx buffer.
 *
 * @param address - The wallet address for the unspent
 * @param chain - The derivation chain
 * @param index - The derivation index
 * @param value - The value in satoshis (bigint)
 * @param network - The network to get the output script
 * @returns WalletUnspent with a valid txid from a wasm-utxo generated transaction
 */
function createWasmWalletUnspent(address, chain, index, value, network) {
    // Get output script from address using correct wasm-utxo function
    const outputScript = wasmUtxo.address.toOutputScriptWithCoin(address, toCoinName(network));
    // Create a mock transaction with output at vout=0
    const { txid } = createMockPrevTx(0, outputScript, BigInt(value));
    return {
        id: `${txid}:0`,
        address,
        chain,
        index,
        value,
    };
}
/**
 * Create a mock previous transaction using wasm-utxo's low-level Psbt class.
 *
 * Creates a valid transaction structure with:
 * - A p2wpkh input (segwit, easy to construct)
 * - Outputs including the target output at the specified vout
 * - Reasonable fee rate (input value slightly higher than output sum)
 *
 * @param vout - The output index where the target output should be placed
 * @param outputScript - The scriptPubKey for the target output
 * @param value - The value in satoshis for the target output
 * @returns Object with prevTx buffer and the transaction's txid
 */
function createMockPrevTx(vout, outputScript, value) {
    // Create a p2wpkh script for the dummy input
    // p2wpkh: OP_0 <20-byte-hash> - we use a deterministic hash
    const dummyHash = (0, sdk_test_1.getSeed)('mock-prev-tx-input').subarray(0, 20);
    const p2wpkhScript = new Uint8Array([0x00, 0x14, ...dummyHash]);
    // Calculate total output value
    const dummyOutputValue = BigInt(1000); // 1000 satoshis per dummy output
    const totalOutputValue = dummyOutputValue * BigInt(vout) + value;
    // Input value = output value + small fee (1000 sat - reasonable fee)
    const inputValue = totalOutputValue + BigInt(1000);
    // Create PSBT using WrapPsbt (exported as Psbt)
    const psbt = new wasmUtxo.Psbt(2, 0);
    // Add dummy segwit input - p2wpkh
    psbt.addInput('01'.repeat(32), // non-zero txid to avoid coinbase detection
    0, inputValue, p2wpkhScript);
    // Add dummy outputs before the target vout
    for (let i = 0; i < vout; i++) {
        psbt.addOutput(p2wpkhScript, dummyOutputValue);
    }
    // Add the target output at the specified vout
    psbt.addOutput(outputScript, value);
    // Get the unsigned transaction bytes
    // Note: This is a valid transaction structure without signatures
    // For prevTx purposes, we just need the output scripts and values
    const txBytes = psbt.getUnsignedTx();
    // Parse with wasm-utxo Transaction to get the txid
    const tx = wasmUtxo.Transaction.fromBytes(txBytes);
    const txid = tx.getId();
    return {
        prevTx: Buffer.from(txBytes),
        txid,
    };
}
/**
 * Create unspent with prevTx using wasm-utxo only (no utxolib dependency).
 * The prevTx is built using wasm-utxo's low-level Psbt class.
 *
 * Key: The unspent id uses the actual txid from the generated prevTx,
 * ensuring the MockRecoveryProvider returns the correct prevTx.
 */
function toUnspentWithPrevTx(input, index, network, rootWalletKeys) {
    const scriptType = input.scriptType === 'taprootKeyPathSpend' ? 'p2trMusig2' : input.scriptType;
    const chain = getInternalChainCode(scriptType);
    const coinName = toCoinName(network);
    // Get the output script for the wallet address
    const outputScript = wasmUtxo.fixedScriptWallet.outputScript(rootWalletKeys, chain, index, coinName);
    // Create mock prevTx with output at vout=0
    const { prevTx, txid } = createMockPrevTx(0, outputScript, input.value);
    // Get the wallet address
    const address = (0, address_1.getWalletAddress)(toCoinName(network), rootWalletKeys, chain, index);
    // Use the actual txid from the prevTx in the unspent id
    return {
        id: `${txid}:0`,
        address,
        chain,
        index,
        value: input.value,
        prevTx,
    };
}
