// @bitgo/abstract-utxo/dist/cjs/test/unit/util/utxoCoins.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.minUtxoCoinNames = exports.utxoCoins = exports.defaultBitGo = void 0;
exports.getNetworkForCoinName = getNetworkForCoinName;
exports.getCoinNameForNetwork = getCoinNameForNetwork;
exports.getMinUtxoCoins = getMinUtxoCoins;
exports.getUtxoCoin = getUtxoCoin;
exports.getUtxoCoinForNetwork = getUtxoCoinForNetwork;
exports.getScriptTypes = getScriptTypes;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../../src");
exports.defaultBitGo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
/** Map coin names to utxolib network objects for test infrastructure */
const coinNameToNetwork = {
    btc: utxolib.networks.bitcoin,
    tbtc: utxolib.networks.testnet,
    tbtcsig: utxolib.networks.bitcoinPublicSignet,
    tbtc4: utxolib.networks.bitcoinTestnet4,
    tbtcbgsig: utxolib.networks.bitcoinBitGoSignet,
    bch: utxolib.networks.bitcoincash,
    tbch: utxolib.networks.bitcoincashTestnet,
    bcha: utxolib.networks.ecash,
    tbcha: utxolib.networks.ecashTest,
    bsv: utxolib.networks.bitcoinsv,
    tbsv: utxolib.networks.bitcoinsvTestnet,
    btg: utxolib.networks.bitcoingold,
    dash: utxolib.networks.dash,
    tdash: utxolib.networks.dashTest,
    doge: utxolib.networks.dogecoin,
    tdoge: utxolib.networks.dogecoinTest,
    ltc: utxolib.networks.litecoin,
    tltc: utxolib.networks.litecoinTest,
    zec: utxolib.networks.zcash,
    tzec: utxolib.networks.zcashTest,
};
/**
 * Get utxolib Network for a coin name. For test infrastructure only.
 */
function getNetworkForCoinName(coinName) {
    const network = coinNameToNetwork[coinName];
    if (!network) {
        throw new Error(`Unknown coin name: ${coinName}`);
    }
    return network;
}
/**
 * Get coin name for a utxolib Network. For test infrastructure only.
 */
function getCoinNameForNetwork(network) {
    for (const [name, n] of Object.entries(coinNameToNetwork)) {
        if (n === network) {
            return name;
        }
    }
    throw new Error(`Unknown network: ${utxolib.getNetworkName(network)}`);
}
const utxoCoinClasses = [
    src_1.Btc,
    src_1.Tbtc,
    src_1.Tbtc4,
    src_1.Tbtcsig,
    src_1.Tbtcbgsig,
    src_1.Bch,
    src_1.Tbch,
    src_1.Bcha,
    src_1.Tbcha,
    src_1.Bsv,
    src_1.Tbsv,
    src_1.Btg,
    src_1.Ltc,
    src_1.Tltc,
    src_1.Dash,
    src_1.Tdash,
    src_1.Doge,
    src_1.Tdoge,
    src_1.Zec,
    src_1.Tzec,
];
function getUtxoCoins(bitgo = exports.defaultBitGo) {
    return utxoCoinClasses
        .map((cls) => {
        try {
            return cls.createInstance(bitgo);
        }
        catch (e) {
            throw new Error(`error creating ${cls.name}: ${e}`);
        }
    })
        .sort((a, b) => {
        const networkList = utxolib.getNetworkList();
        return networkList.indexOf(getNetworkForCoinName(a.name)) - networkList.indexOf(getNetworkForCoinName(b.name));
    });
}
exports.utxoCoins = getUtxoCoins();
/**
 * Minimal subset of coins for comprehensive test coverage.
 * - btc: Full feature set (segwit, taproot, musig2)
 * - bch: replay protection addresses
 * - zec: special transaction format (Overwinter/Sapling)
 */
exports.minUtxoCoinNames = ['btc', 'bch', 'zec'];
/**
 * Get minimal set of coins for testing. Covers ~99% of feature paths.
 */
function getMinUtxoCoins() {
    return exports.minUtxoCoinNames.map(getUtxoCoin);
}
function getUtxoCoin(name) {
    for (const c of exports.utxoCoins) {
        if (c.getChain() === name) {
            return c;
        }
    }
    throw new Error(`no coin with name ${name}`);
}
function getUtxoCoinForNetwork(n) {
    for (const c of exports.utxoCoins) {
        if (getNetworkForCoinName(c.name) === n) {
            return c;
        }
    }
    throw new Error(`no coin for network ${utxolib.getNetworkName(n)}`);
}
function getScriptTypes(coin, txFormat) {
    return ['p2shP2pk', 'p2sh', 'p2shP2wsh', 'p2wsh', 'p2tr', 'p2trMusig2', 'taprootKeyPathSpend'].filter((t) => {
        if (t === 'p2shP2pk') {
            return (0, src_1.getReplayProtectionAddresses)(coin.name).length > 0;
        }
        if (txFormat === 'legacy') {
            if (t === 'p2tr' || t === 'p2trMusig2' || t === 'taprootKeyPathSpend') {
                return false;
            }
        }
        if (t === 'taprootKeyPathSpend') {
            return coin.supportsAddressType('p2trMusig2');
        }
        return coin.supportsAddressType(t);
    });
}
