// @bitgo/abstract-utxo/dist/cjs/src/descriptor/createWallet/createDescriptorWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDescriptorWallet = createDescriptorWallet;
exports.createDescriptorWalletWithWalletPassphrase = createDescriptorWalletWithWalletPassphrase;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const sdk_core_1 = require("@bitgo/sdk-core");
async function createDescriptorWallet(bitgo, coin, { descriptors, ...params }) {
    // We don't use `coin.wallets().add` here because it does a bunch of validation that does not make sense
    // for descriptor wallets.
    const newWallet = await bitgo
        .post(coin.url('/wallet/add'))
        .send({
        ...params,
        coinSpecific: { descriptors },
    })
        .result();
    return new sdk_core_1.Wallet(bitgo, coin, newWallet);
}
async function createDescriptorWalletWithWalletPassphrase(bitgo, coin, { enterprise, walletPassphrase, descriptorsFromKeys, ...params }) {
    const userKeychain = await coin.keychains().createUserKeychain(walletPassphrase);
    const backupKeychain = await coin.keychains().createBackup();
    const bitgoKeychain = await coin.keychains().createBitGo({ enterprise });
    if (!userKeychain.prv) {
        throw new Error('Missing private key');
    }
    const userKey = wasm_utxo_1.BIP32.fromBase58(userKeychain.prv);
    const cosigners = [backupKeychain, bitgoKeychain].map((keychain) => {
        if (!keychain.pub) {
            throw new Error('Missing public key');
        }
        return wasm_utxo_1.BIP32.fromBase58(keychain.pub);
    });
    return createDescriptorWallet(bitgo, coin, {
        ...params,
        type: 'hot',
        enterprise,
        keys: [userKeychain.id, backupKeychain.id, bitgoKeychain.id],
        descriptors: descriptorsFromKeys(userKey, cosigners),
    });
}
