// @bitgo/wasm-utxo/dist/cjs/js/inspect/index.js
"use strict";
/**
 * Inspect - TypeScript bindings for PSBT and Transaction parsing
 *
 * Provides typed wrappers around the WASM inspect functions that return
 * hierarchical node structures suitable for display as collapsible trees.
 *
 * Import via: `import { ... } from "@bitgo/wasm-utxo/inspect"`
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.allNetworks = void 0;
exports.parsePsbtToNode = parsePsbtToNode;
exports.parseTxToNode = parseTxToNode;
exports.tryParsePsbt = tryParsePsbt;
exports.tryParseTx = tryParseTx;
exports.parsePsbtRawToNode = parsePsbtRawToNode;
exports.tryParsePsbtRaw = tryParsePsbtRaw;
exports.isInspectEnabled = isInspectEnabled;
const wasm_utxo_js_1 = require("../wasm/wasm_utxo.js");
/** All supported networks in order of parsing priority */
exports.allNetworks = [
    "btc",
    "tbtc",
    "tbtc4",
    "tbtcsig",
    "tbtcbgsig",
    "ltc",
    "tltc",
    "bch",
    "tbch",
    "bcha",
    "tbcha",
    "btg",
    "tbtg",
    "bsv",
    "tbsv",
    "dash",
    "tdash",
    "doge",
    "tdoge",
    "zec",
    "tzec",
];
/**
 * Parse a PSBT and return a typed node tree.
 *
 * @param psbtBytes - The raw PSBT bytes
 * @param network - The network coin name (e.g., "btc", "ltc", "bch")
 * @returns A Node tree representing the parsed PSBT structure
 * @throws If the PSBT bytes are invalid or network is unknown
 */
function parsePsbtToNode(psbtBytes, network) {
    const json = (0, wasm_utxo_js_1.parsePsbtToJson)(psbtBytes, network);
    return JSON.parse(json);
}
/**
 * Parse a transaction and return a typed node tree.
 *
 * @param txBytes - The raw transaction bytes
 * @param network - The network coin name (e.g., "btc", "ltc", "bch")
 * @returns A Node tree representing the parsed transaction structure
 * @throws If the transaction bytes are invalid or network is unknown
 */
function parseTxToNode(txBytes, network) {
    const json = (0, wasm_utxo_js_1.parseTxToJson)(txBytes, network);
    return JSON.parse(json);
}
/**
 * Try to parse a PSBT with all networks and return the first one that succeeds.
 *
 * @param psbtBytes - The raw PSBT bytes
 * @param networks - Optional list of networks to try (defaults to all networks)
 * @returns An object with the parsed Node and detected network, or null if all fail
 */
function tryParsePsbt(psbtBytes, networks = exports.allNetworks) {
    for (const network of networks) {
        try {
            const node = parsePsbtToNode(psbtBytes, network);
            return { node, network };
        }
        catch {
            // Try next network
        }
    }
    return null;
}
/**
 * Try to parse a transaction with all networks and return the first one that succeeds.
 *
 * @param txBytes - The raw transaction bytes
 * @param networks - Optional list of networks to try (defaults to all networks)
 * @returns An object with the parsed Node and detected network, or null if all fail
 */
function tryParseTx(txBytes, networks = exports.allNetworks) {
    for (const network of networks) {
        try {
            const node = parseTxToNode(txBytes, network);
            return { node, network };
        }
        catch {
            // Try next network
        }
    }
    return null;
}
/**
 * Parse a PSBT at the raw byte level and return a typed node tree.
 *
 * Unlike `parsePsbtToNode`, this function exposes the raw key-value pair
 * structure as defined in BIP-174, showing:
 * - Raw key type IDs and their human-readable names
 * - Proprietary keys with their structured format
 * - Unknown/unrecognized keys that standard parsers might skip
 *
 * @param psbtBytes - The raw PSBT bytes
 * @param network - The network coin name (e.g., "btc", "ltc", "zec")
 * @returns A Node tree representing the raw PSBT key-value structure
 * @throws If the PSBT bytes are invalid or network is unknown
 */
function parsePsbtRawToNode(psbtBytes, network) {
    const json = (0, wasm_utxo_js_1.parsePsbtRawToJson)(psbtBytes, network);
    return JSON.parse(json);
}
/**
 * Try to parse a raw PSBT with all networks and return the first one that succeeds.
 *
 * @param psbtBytes - The raw PSBT bytes
 * @param networks - Optional list of networks to try (defaults to all networks)
 * @returns An object with the parsed Node and detected network, or null if all fail
 */
function tryParsePsbtRaw(psbtBytes, networks = exports.allNetworks) {
    for (const network of networks) {
        try {
            const node = parsePsbtRawToNode(psbtBytes, network);
            return { node, network };
        }
        catch {
            // Try next network
        }
    }
    return null;
}
/**
 * Check if the inspect feature is enabled in the WASM build.
 *
 * @returns true if the feature is enabled, false otherwise
 */
function isInspectEnabled() {
    return (0, wasm_utxo_js_1.isInspectEnabled)();
}
