// @bitgo/sdk-coin-btc/dist/cjs/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = exports.Tbtcbgsig = exports.Tbtc4 = exports.Tbtcsig = exports.Tbtc = exports.InscriptionBuilder = exports.Btc = void 0;
const abstract_utxo_1 = require("@bitgo/abstract-utxo");
var abstract_utxo_2 = require("@bitgo/abstract-utxo");
Object.defineProperty(exports, "Btc", { enumerable: true, get: function () { return abstract_utxo_2.Btc; } });
Object.defineProperty(exports, "InscriptionBuilder", { enumerable: true, get: function () { return abstract_utxo_2.InscriptionBuilder; } });
var abstract_utxo_3 = require("@bitgo/abstract-utxo");
Object.defineProperty(exports, "Tbtc", { enumerable: true, get: function () { return abstract_utxo_3.Tbtc; } });
var abstract_utxo_4 = require("@bitgo/abstract-utxo");
Object.defineProperty(exports, "Tbtcsig", { enumerable: true, get: function () { return abstract_utxo_4.Tbtcsig; } });
var abstract_utxo_5 = require("@bitgo/abstract-utxo");
Object.defineProperty(exports, "Tbtc4", { enumerable: true, get: function () { return abstract_utxo_5.Tbtc4; } });
var abstract_utxo_6 = require("@bitgo/abstract-utxo");
Object.defineProperty(exports, "Tbtcbgsig", { enumerable: true, get: function () { return abstract_utxo_6.Tbtcbgsig; } });
const register = (sdk) => {
    sdk.register('btc', abstract_utxo_1.Btc.createInstance);
    sdk.register('tbtc', abstract_utxo_1.Tbtc.createInstance);
    sdk.register('tbtcsig', abstract_utxo_1.Tbtcsig.createInstance);
    sdk.register('tbtcbgsig', abstract_utxo_1.Tbtcbgsig.createInstance);
    sdk.register('tbtc4', abstract_utxo_1.Tbtc4.createInstance);
};
exports.register = register;
