// @bitgo/sdk-coin-avaxp/dist/test/resources/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERROR_CHAIN_ID_INVALID_CHECKSUM = exports.ERROR_CHAIN_ID_NOT_BASE58 = exports.ERROR_NONCE = exports.ERROR_AMOUNT = exports.ERROR_KEY_CANNOT_SIGN = exports.ERROR_CHAIN_ID_LENGTH = exports.ERROR_UTXOS_AMOUNT = exports.ERROR_UTXOS_EMPTY = exports.ERROR_WALLET_INITIALIZATION = exports.ERROR_DELEGATION_FEE = exports.ERROR_STAKE_AMOUNT = exports.ERROR_STAKE_DURATION_LONG_TIME = exports.ERROR_STAKE_DURATION_SHORT_TIME = exports.ERROR_STAKE_START_TIME_TOO_SHORT = exports.ERROR_RAW_PARSING = exports.ERROR_EMPTY_RAW_TRANSACTION = exports.ERROR_NODE_ID_LENGTH = exports.ERROR_NODE_ID = void 0;
exports.ERROR_NODE_ID = 'Invalid transaction: invalid NodeID tag';
exports.ERROR_NODE_ID_LENGTH = 'Invalid transaction: NodeID is not in cb58 format';
exports.ERROR_EMPTY_RAW_TRANSACTION = 'Raw transaction is empty';
exports.ERROR_RAW_PARSING = 'Raw transaction is not hex string';
exports.ERROR_STAKE_START_TIME_TOO_SHORT = 'Start time needs to be one day greater than current time';
exports.ERROR_STAKE_DURATION_SHORT_TIME = 'End date must be greater than or equal to two weeks';
exports.ERROR_STAKE_DURATION_LONG_TIME = 'End date must be less than or equal to one year';
exports.ERROR_STAKE_AMOUNT = 'Minimum staking amount is 2000 AVAX.';
exports.ERROR_DELEGATION_FEE = 'Delegation fee cannot be less than 2';
exports.ERROR_WALLET_INITIALIZATION = 'Wallet initialization is not needed';
exports.ERROR_UTXOS_EMPTY = "Utxos can't be empty array";
exports.ERROR_UTXOS_AMOUNT = 'Utxos required amount';
exports.ERROR_CHAIN_ID_LENGTH = 'Chain id are 32 byte size';
exports.ERROR_KEY_CANNOT_SIGN = 'Private key cannot sign the transaction';
exports.ERROR_AMOUNT = 'Amount must be greater than 0';
exports.ERROR_NONCE = 'Nonce must be greater or equal than 0';
exports.ERROR_CHAIN_ID_NOT_BASE58 = 'Error - Base58.decode: not a valid base58 string';
exports.ERROR_CHAIN_ID_INVALID_CHECKSUM = 'Error - BinTools.cb58Decode: invalid checksum';
