// @bitgo/sdk-coin-avaxp/dist/test/resources/keychains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keychains = void 0;
exports.keychains = [
    {
        id: '624f0dcc93cbcc0008d88df2369a565e',
        pub: 'xpub661MyMwAqRbcEeRkBciuaUfF4C1jgBcnj2RXdnt9gokx4CFRBUp4bsbk5hXHC1BrBDZLDNecVsUCMmoLpPhWdPZhPiTsHSoxNoGVW9KtiEQ',
        ethAddress: '0xcfbf38770af3a95da7998537a481434e2cb9b2fa',
        source: 'user',
        type: 'independent',
        encryptedPrv: '{"iv":"Z2XySTRNipFZ06/EXynwvA==","v":1,"iter":10000,"ks":256,"ts":64,"mode":"ccm","adata":"","cipher":"aes","salt":"KGRPbZ2jt1g=","ct":"szpCbDLFIlRZvCBV60SWBEMYXvny7YlBtu4ffjlctDQGjR4/+vfCkovgGHs+Xvf/eIlUM3Kicubg+Sdp61MImjMT/umZ3IJT1E2I9mM0QDqpzXlohTGnJ4vgfHgCz3QkB4uYm5mqaD4LtRbvZbGhGrc5jzrLzqQ="}',
    },
    {
        id: '624f0dcd93cbcc0008d88e0fc4261a38',
        pub: 'xpub661MyMwAqRbcGeqZVFgQfcD8zLoxaZL7y4cVAjhE8ybMTpvbppP6rc22a69BgcNVo74yL8fWPzNM5vAozBE7chzGYoPDJMyJ39F2HeAsGcn',
        ethAddress: '0xbf37f39208d77e3254b7efbcab1432b9c353e337',
        source: 'backup',
        type: 'independent',
        encryptedPrv: '{"iv":"T9gdJnSAEWFsLZ4cg9VA8g==","v":1,"iter":10000,"ks":256,"ts":64,"mode":"ccm","adata":"","cipher":"aes","salt":"FaLlns3mPiI=","ct":"QW5Zq9qJoDxDrK60zTAM6Lg+S4KP9FcEn9AHw5UIyakSBlD0XjVTluZ9PlTABjIlp9cQvMef/SH8Em1d4ash0PACoqBz2IxPwhW9h6uyQBdqk97iPrnM2rOQobsy9p0ILJM10fOgB+EEFYX5yQ5gyfEcK060j/Q="}',
    },
    {
        id: '624f0dce10610a0007fc5282353187ae',
        pub: 'xpub661MyMwAqRbcFVMAYJe51sgXaiFLeUb1v4u3B63CgBNMmMjtWBo32AS3bunsBUZMdi37pzovtEg5mVf6wBKayTYapGQRxymQjcmHaVmSPz8',
        ethAddress: '0x7527720b5638d2f5e2b272b20fc96d2223528d0e',
        source: 'bitgo',
        type: 'independent',
        isBitGo: true,
    },
];
