// @bitgo/utxo-core/dist/cjs/src/testutil/fixtures.utils.js
"use strict";
/**
 * Contains helpers for working with test fixtures
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFixture = getFixture;
exports.jsonNormalize = jsonNormalize;
const fs = __importStar(require("fs"));
const mpath = __importStar(require("path"));
function isNodeJsError(e) {
    return e instanceof Error && typeof e.code === 'string';
}
function fixtureEncoding(path) {
    if (path.endsWith('.json')) {
        return 'json';
    }
    if (path.endsWith('.hex')) {
        return 'hex';
    }
    if (path.endsWith('.txt')) {
        return 'txt';
    }
    throw new Error(`unknown fixture encoding for ${path}`);
}
function decodeFixture(raw, encoding) {
    switch (encoding) {
        case 'json':
            return JSON.parse(raw);
        case 'hex':
            return Buffer.from(raw, 'hex');
        case 'txt':
            return raw;
    }
}
function encodeFixture(value, encoding) {
    switch (encoding) {
        case 'json':
            return JSON.stringify(value, null, 2) + '\n';
        case 'hex':
            if (!Buffer.isBuffer(value)) {
                throw new Error(`expected Buffer, got ${typeof value}`);
            }
            return value.toString('hex');
        case 'txt':
            if (typeof value !== 'string') {
                throw new Error(`expected string, got ${typeof value}`);
            }
            return value;
    }
}
/**
 * Return fixture described in `path`.
 *
 * If file does not exist and `defaultValue` is provided, writes defaultValue to `path` and throws an error.
 *
 * @param path
 * @param defaultValue
 * @return T - fixture content
 */
async function getFixture(path, defaultValue) {
    try {
        await fs.promises.stat(mpath.dirname(path));
    }
    catch (e) {
        if (isNodeJsError(e) && e.code === 'ENOENT') {
            throw new Error(`fixture directory ${mpath.dirname(path)} not found, please create it first`);
        }
        throw e;
    }
    const encoding = fixtureEncoding(path);
    try {
        return decodeFixture(await fs.promises.readFile(path, 'utf8'), encoding);
    }
    catch (e) {
        if (isNodeJsError(e) && e.code === 'ENOENT') {
            if (process.env.WRITE_FIXTURES === '0') {
                throw new Error(`fixture ${path} not found, WRITE_FIXTURES=0`);
            }
            if (defaultValue === undefined) {
                throw new Error(`fixture ${path} not found and no default value given`);
            }
            if (typeof defaultValue === 'function') {
                defaultValue = await defaultValue();
            }
            await fs.promises.writeFile(path, encodeFixture(defaultValue, encoding));
            throw new Error(`wrote default value for ${path}, please inspect and restart test`);
        }
        throw e;
    }
}
function jsonNormalize(v) {
    return JSON.parse(JSON.stringify(v));
}
