// @bitgo/utxo-core/dist/cjs/src/testutil/toPlainObject.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.matchPath = matchPath;
exports.toPlainObject = toPlainObject;
function matchPath(a, b) {
    return a.length === b.length && a.every((e, i) => e === b[i]);
}
function includePath(opts, path) {
    if (!opts.ignorePaths) {
        return true;
    }
    if (typeof opts.ignorePaths === 'function') {
        return !opts.ignorePaths(path);
    }
    return !opts.ignorePaths.some((ignorePath) => matchPath(path, ignorePath.split('.')));
}
function toPlainEntries(key, value, opts, path) {
    if (!includePath(opts, [...path, key])) {
        return [];
    }
    if (value === undefined && (opts.skipUndefinedValues ?? true)) {
        return [];
    }
    return [[key, toPlainObject(value, opts, [...path, key])]];
}
function getAllDescriptors(v) {
    if (v === null || typeof v !== 'object') {
        return {};
    }
    const descriptors = Object.getOwnPropertyDescriptors(v);
    const proto = Object.getPrototypeOf(v);
    if (proto) {
        Object.assign(descriptors, getAllDescriptors(proto));
    }
    return descriptors;
}
function toPlainObjectFromPropertyDescriptors(v, opts, path) {
    const descriptors = getAllDescriptors(v);
    return Object.fromEntries(Object.entries(descriptors).flatMap(([key, descriptor]) => {
        if (typeof descriptor.value === 'function') {
            return [];
        }
        if (descriptor.value !== undefined) {
            return toPlainEntries(key, descriptor.value, opts, path);
        }
        if (typeof descriptor.get === 'function') {
            return toPlainEntries(key, descriptor.get.call(v), opts, path);
        }
        return [];
    }));
}
function toPlainObject(v, opts = {}, path = []) {
    if (opts.apply) {
        const result = opts.apply(v, path);
        if (result !== undefined) {
            return result;
        }
    }
    switch (typeof v) {
        case 'string':
        case 'number':
        case 'boolean':
        case 'undefined':
            return v;
        case 'bigint':
            return v.toString();
        case 'function':
        case 'symbol':
            return undefined;
    }
    if (v === null) {
        return v;
    }
    if (Buffer.isBuffer(v)) {
        return v.toString('hex');
    }
    if (Array.isArray(v)) {
        return v.map((e, i) => toPlainObject(e, opts, [...path, i]));
    }
    if (typeof v === 'object') {
        const result = Object.fromEntries(Object.entries(v).flatMap(([key, value]) => toPlainEntries(key, value, opts, path)));
        if (opts.propertyDescriptors) {
            Object.assign(result, toPlainObjectFromPropertyDescriptors(v, opts, path));
        }
        return result;
    }
    throw new Error(`unknown v ${typeof v}`);
}
