// @bitgo/abstract-lightning/dist/test/unit/lightning/codecs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const E = __importStar(require("fp-ts/Either"));
const function_1 = require("fp-ts/function");
const codecs_1 = require("../../../src/codecs");
function describeCodec(c, valid, invalid) {
    describe('Codec ' + c.name, function () {
        it('parses valid inputs', function () {
            valid.forEach((v) => {
                assert_1.default.strictEqual(E.isRight(c.decode(v)), true);
            });
        });
        it('rejects invalid inputs', function () {
            invalid.forEach((v) => {
                assert_1.default.strictEqual(E.isRight(c.decode(v)), false);
            });
        });
        it('encode ∘ decode', function () {
            valid.forEach((v) => {
                const roundTripped = (0, function_1.pipe)(v, c.decode, E.map(c.encode));
                assert_1.default.deepStrictEqual(roundTripped, E.right(v));
            });
        });
    });
}
describe('Codecs', function () {
    describeCodec(codecs_1.LightningKeychain, [
        {
            id: 'id',
            pub: 'xpub',
            encryptedPrv: 'encryptedPrv',
            source: 'user',
        },
        {
            id: 'id',
            pub: 'xpub',
            source: 'user',
        },
    ], [
        null,
        'abg',
        1,
        {
            id: 'id',
            pub: 'xpub',
            encryptedPrv: 'encryptedPrv',
            source: 'backup',
        },
    ]);
    describeCodec(codecs_1.LightningAuthKeychain, [
        {
            id: 'id',
            pub: 'xpub',
            encryptedPrv: 'encryptedPrv',
            source: 'user',
            coinSpecific: {
                lnbtc: {
                    purpose: 'userAuth',
                },
            },
        },
        {
            id: 'id',
            pub: 'xpub',
            source: 'user',
            coinSpecific: {
                lnbtc: {
                    purpose: 'userAuth',
                },
            },
        },
    ], [
        null,
        'abg',
        1,
        {
            id: 'id',
            pub: 'xpub',
            encryptedPrv: 'encryptedPrv',
            source: 'user',
            coinSpecific: {
                lnbtc: {
                    purpose: 'dummy',
                },
            },
        },
        {
            id: 'id',
            pub: 'xpub',
            encryptedPrv: 'encryptedPrv',
        },
    ]);
    describeCodec(codecs_1.UpdateLightningWalletClientRequest, [
        {
            signerAdminMacaroon: 'signerAdminMacaroon',
            signerHost: '127.0.0.1',
            signerTlsCert: 'signerTlsCert',
            signerTlsKey: 'signerTlsKey',
            watchOnly: {
                master_key_birthday_timestamp: 'master_key_birthday_timestamp',
                master_key_fingerprint: 'master_key_fingerprint',
                accounts: [{ purpose: 1, coin_type: 1, account: 1, xpub: 'xpub' }],
            },
            signerMacaroon: 'signerMacaroon',
            passphrase: 'passphrase',
        },
        {
            signerAdminMacaroon: 'signerAdminMacaroon',
            passphrase: 'passphrase',
        },
        { passphrase: 'passphrase' },
    ], [null, 'abg', 1]);
});
