// @bitgo/logger/dist/src/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
const sanitizeLog_1 = require("./sanitizeLog");
/**
 * BitGo Logger with automatic sanitization for all environments
 */
class BitGoLogger {
    /**
     * Sanitizes arguments for all environments
     */
    sanitizeArgs(args) {
        return args.map((arg) => {
            if (arg instanceof Error) {
                return (0, sanitizeLog_1.sanitize)((0, sanitizeLog_1.getErrorData)(arg));
            }
            return (0, sanitizeLog_1.sanitize)(arg);
        });
    }
    /**
     * Logs a message with automatic sanitization
     */
    log(...args) {
        const sanitized = this.sanitizeArgs(args);
        // eslint-disable-next-line no-console
        console.log(...sanitized);
    }
    /**
     * Logs an error message with automatic sanitization
     */
    error(...args) {
        const sanitized = this.sanitizeArgs(args);
        // eslint-disable-next-line no-console
        console.error(...sanitized);
    }
    /**
     * Logs a warning message with automatic sanitization
     */
    warn(...args) {
        const sanitized = this.sanitizeArgs(args);
        // eslint-disable-next-line no-console
        console.warn(...sanitized);
    }
    /**
     * Logs an info message with automatic sanitization
     */
    info(...args) {
        const sanitized = this.sanitizeArgs(args);
        // eslint-disable-next-line no-console
        console.info(...sanitized);
    }
}
/**
 * Singleton logger instance
 */
exports.logger = new BitGoLogger();
