// @bitgo/sdk-coin-xlm/dist/src/lib/keyPair.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
const stellar = __importStar(require("stellar-sdk"));
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("./utils");
class KeyPair extends sdk_core_1.Ed25519KeyPair {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    constructor(source) {
        super(source);
    }
    getKeys(raw = false) {
        const publicKey = (0, utils_1.encodePublicKey)(Buffer.from(this.keyPair.pub, 'hex'));
        const result = {
            pub: raw ? (0, utils_1.decodePublicKey)(publicKey).toString('hex') : publicKey,
        };
        if (this.keyPair.prv) {
            const privateKey = (0, utils_1.encodePrivateKey)(Buffer.from(this.keyPair.prv, 'hex'));
            result.prv = raw ? (0, utils_1.decodePrivateKey)(privateKey).toString('hex') : privateKey;
        }
        return result;
    }
    /** @inheritdoc */
    getAddress() {
        return (0, utils_1.encodePublicKey)(Buffer.from(this.keyPair.pub, 'hex'));
    }
    /** @inheritdoc */
    recordKeysFromPublicKeyInProtocolFormat(pub) {
        try {
            const publicKey = (0, utils_1.decodePublicKey)(pub);
            return { pub: publicKey.toString('hex') };
        }
        catch (e) {
            throw new sdk_core_1.InvalidKey('Invalid public key: ' + pub);
        }
    }
    /** @inheritdoc */
    recordKeysFromPrivateKeyInProtocolFormat(prv) {
        try {
            const kp = stellar.Keypair.fromSecret(prv);
            const publicKey = (0, utils_1.decodePublicKey)(kp.publicKey());
            const privateKey = (0, utils_1.decodePrivateKey)(kp.secret());
            return {
                pub: publicKey.toString('hex'),
                prv: privateKey.toString('hex'),
            };
        }
        catch (e) {
            throw new sdk_core_1.InvalidKey('Invalid private key: ' + prv);
        }
    }
}
exports.KeyPair = KeyPair;
