// @bitgo/sdk-coin-ton/dist/src/jettonToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JettonToken = void 0;
const statics_1 = require("@bitgo/statics");
const ton_1 = require("./ton");
class JettonToken extends ton_1.Ton {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('ton') : statics_1.coins.get('tton');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new JettonToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfig = [...statics_1.tokens.bitcoin.ton.tokens, ...statics_1.tokens.testnet.ton.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfig) {
            const tokenConstructor = JettonToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get contractAddress() {
        return this.tokenConfig.contractAddress;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Ton Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.JettonToken = JettonToken;
