// @bitgo/sdk-coin-ton/dist/src/ton.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ton = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const _ = __importStar(require("lodash"));
const tonweb_1 = __importDefault(require("tonweb"));
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const statics_1 = require("@bitgo/statics");
const keyPair_1 = require("./lib/keyPair");
const lib_1 = require("./lib");
const utils_1 = require("./lib/utils");
class Ton extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Ton(bitgo, staticsCoin);
    }
    /**
     * Factor between the coin's base unit and its smallest subdivison
     */
    getBaseFactor() {
        return 1e9;
    }
    getChain() {
        return 'ton';
    }
    getFamily() {
        return 'ton';
    }
    getFullName() {
        return 'Ton';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    getMPCAlgorithm() {
        return 'eddsa';
    }
    allowsAccountConsolidations() {
        return true;
    }
    getAddressDetails(address) {
        const addressComponents = address.split('?memoId=');
        if (addressComponents.length > 1) {
            return {
                address: addressComponents[0],
                memoId: addressComponents[1],
            };
        }
        else {
            return {
                address: addressComponents[0],
            };
        }
    }
    async verifyTransaction(params) {
        const { txPrebuild: txPrebuild, txParams: txParams } = params;
        const rawTx = txPrebuild.txHex;
        if (!rawTx) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const txBuilder = this.getBuilder().from(Buffer.from(rawTx, 'hex').toString('base64'));
        const transaction = await txBuilder.build();
        const explainedTx = transaction.explainTransaction();
        if (txParams.recipients !== undefined) {
            const filteredRecipients = txParams.recipients?.map((recipient) => {
                const destination = this.getAddressDetails(recipient.address);
                return {
                    address: new tonweb_1.default.Address(destination.address).toString(true, true, true),
                    amount: BigInt(recipient.amount),
                };
            });
            const filteredOutputs = explainedTx.outputs.map((output) => {
                return {
                    address: new tonweb_1.default.Address(output.address).toString(true, true, true),
                    amount: BigInt(output.amount),
                };
            });
            if (!_.isEqual(filteredOutputs, filteredRecipients)) {
                throw new Error('Tx outputs does not match with expected txParams recipients');
            }
            let totalAmount = new bignumber_js_1.default(0);
            for (const recipients of txParams.recipients) {
                totalAmount = totalAmount.plus(recipients.amount);
            }
            if (!totalAmount.isEqualTo(explainedTx.outputAmount)) {
                throw new Error('Tx total amount does not match with expected total amount field');
            }
        }
        return true;
    }
    async isWalletAddress(params) {
        const { keychains, address: newAddress, index } = params;
        if (!this.isValidAddress(newAddress)) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${newAddress}`);
        }
        const [address, memoId] = newAddress.split('?memoId=');
        // TON supports memoId for address tagging - verify it matches the index
        if (memoId) {
            return memoId === `${index}`;
        }
        const commonKeychain = (0, sdk_core_1.extractCommonKeychain)(keychains);
        const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
        const derivationPath = 'm/' + index;
        const derivedPublicKey = MPC.deriveUnhardened(commonKeychain, derivationPath).slice(0, 64);
        const expectedAddress = await lib_1.Utils.default.getAddressFromPublicKey(derivedPublicKey);
        return address === expectedAddress;
    }
    async parseTransaction(params) {
        const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
        const transactionBuilder = factory.from(Buffer.from(params.txHex, 'hex').toString('base64'));
        if (typeof params.toAddressBounceable === 'boolean') {
            transactionBuilder.toAddressBounceable(params.toAddressBounceable);
        }
        if (typeof params.fromAddressBounceable === 'boolean') {
            transactionBuilder.fromAddressBounceable(params.fromAddressBounceable);
        }
        const rebuiltTransaction = await transactionBuilder.build();
        const parsedTransaction = rebuiltTransaction.toJson();
        return {
            inputs: [
                {
                    address: parsedTransaction.sender,
                    amount: parsedTransaction.amount,
                },
            ],
            outputs: [
                {
                    address: parsedTransaction.destination,
                    amount: parsedTransaction.amount,
                },
            ],
        };
    }
    generateKeyPair(seed) {
        const keyPair = seed ? new keyPair_1.KeyPair({ seed }) : new keyPair_1.KeyPair();
        const keys = keyPair.getKeys();
        if (!keys.prv) {
            throw new Error('Missing prv in key generation.');
        }
        return {
            pub: keys.pub,
            prv: keys.prv,
        };
    }
    isValidPub(pub) {
        throw new Error('Method not implemented.');
    }
    isValidAddress(address) {
        try {
            const addressBase64 = address.replace(/\+/g, '-').replace(/\//g, '_');
            const buf = Buffer.from(addressBase64.split('?memoId=')[0], 'base64');
            return buf.length === 36;
        }
        catch {
            return false;
        }
    }
    signTransaction(params) {
        throw new Error('Method not implemented.');
    }
    /** @inheritDoc */
    async getSignablePayload(serializedTx) {
        const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
        const rebuiltTransaction = await factory.from(serializedTx).build();
        return rebuiltTransaction.signablePayload;
    }
    /** @inheritDoc */
    async explainTransaction(params) {
        try {
            const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
            const transactionBuilder = factory.from(Buffer.from(params.txHex, 'hex').toString('base64'));
            const { toAddressBounceable, fromAddressBounceable } = params;
            if (typeof toAddressBounceable === 'boolean') {
                transactionBuilder.toAddressBounceable(toAddressBounceable);
            }
            if (typeof fromAddressBounceable === 'boolean') {
                transactionBuilder.fromAddressBounceable(fromAddressBounceable);
            }
            const rebuiltTransaction = await transactionBuilder.build();
            return rebuiltTransaction.explainTransaction();
        }
        catch {
            throw new Error('Invalid transaction');
        }
    }
    getPublicNodeUrl() {
        return sdk_core_1.Environments[this.bitgo.getEnv()].tonNodeUrl;
    }
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    async recover(params) {
        if (!params.bitgoKey) {
            throw new Error('missing bitgoKey');
        }
        if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
            throw new Error('invalid recoveryDestination');
        }
        if (!params.apiKey) {
            throw new Error('missing apiKey');
        }
        const bitgoKey = params.bitgoKey.replace(/\s/g, '');
        const isUnsignedSweep = !params.userKey && !params.backupKey && !params.walletPassphrase;
        // Build the transaction
        const tonweb = new tonweb_1.default(new tonweb_1.default.HttpProvider(this.getPublicNodeUrl(), { apiKey: params.apiKey }));
        const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
        const index = params.index || 0;
        const currPath = params.seed ? (0, sdk_lib_mpc_1.getDerivationPath)(params.seed) + `/${index}` : `m/${index}`;
        const accountId = MPC.deriveUnhardened(bitgoKey, currPath).slice(0, 64);
        const senderAddr = await lib_1.Utils.default.getAddressFromPublicKey(accountId);
        const balance = await tonweb.getBalance(senderAddr);
        const jettonBalances = [];
        if (params.senderJettonAddress) {
            try {
                const jettonWalletData = await tonweb.provider.call(params.senderJettonAddress, 'get_wallet_data');
                const jettonBalance = jettonWalletData.stack[0][1];
                if (jettonBalance && new bignumber_js_1.default(jettonBalance).gt(0)) {
                    jettonBalances.push({
                        walletAddress: params.senderJettonAddress,
                        balance: jettonBalance,
                    });
                }
            }
            catch (e) {
                throw new Error(`Failed to query jetton balance for address ${params.senderJettonAddress}: ${e.message}`);
            }
        }
        const WalletClass = tonweb.wallet.all['v4R2'];
        const wallet = new WalletClass(tonweb.provider, {
            publicKey: tonweb.utils.hexToBytes(accountId),
            wc: 0,
        });
        const seqnoResult = await wallet.methods.seqno().call();
        const seqno = seqnoResult !== null && seqnoResult !== undefined ? seqnoResult : 0;
        const factory = this.getBuilder();
        const expireAt = Math.floor(Date.now() / 1e3) + 60 * 60 * 24 * 7;
        let txBuilder;
        let unsignedTransaction;
        let feeEstimate;
        let transactionType;
        if ((params.jettonMaster || params.senderJettonAddress) && jettonBalances.length > 0) {
            const jettonInfo = jettonBalances[0];
            const tonAmount = '50000000';
            const forwardTonAmount = '1';
            const totalRequiredTon = new bignumber_js_1.default(tonAmount).plus(new bignumber_js_1.default(forwardTonAmount));
            if (new bignumber_js_1.default(balance).lt(totalRequiredTon)) {
                throw new Error(`Insufficient TON balance for jetton transfer. Required: ${totalRequiredTon.toString()} nanoTON, Available: ${balance}`);
            }
            txBuilder = factory
                .getTokenTransferBuilder()
                .sender(senderAddr)
                .sequenceNumber(seqno)
                .publicKey(accountId)
                .expireTime(expireAt);
            txBuilder.recipient(params.recoveryDestination, jettonInfo.walletAddress, tonAmount, jettonInfo.balance, forwardTonAmount);
            unsignedTransaction = await txBuilder.build();
            feeEstimate = parseInt(tonAmount, 10);
            transactionType = 'jetton';
        }
        else {
            if (new bignumber_js_1.default(balance).isEqualTo(0)) {
                throw Error('Did not find address with TON balance to recover');
            }
            const tonFeeEstimate = await (0, utils_1.getFeeEstimate)(wallet, params.recoveryDestination, balance, seqno);
            const totalFeeEstimate = Math.round((tonFeeEstimate.source_fees.in_fwd_fee +
                tonFeeEstimate.source_fees.storage_fee +
                tonFeeEstimate.source_fees.gas_fee +
                tonFeeEstimate.source_fees.fwd_fee) *
                1.5);
            if (new bignumber_js_1.default(totalFeeEstimate).gte(balance)) {
                throw new Error(`Insufficient TON balance for transaction. Required: ${totalFeeEstimate} nanoTON, Available: ${balance}`);
            }
            txBuilder = factory
                .getTransferBuilder()
                .sender(senderAddr)
                .sequenceNumber(seqno)
                .publicKey(accountId)
                .expireTime(expireAt);
            txBuilder.send({
                address: params.recoveryDestination,
                amount: new bignumber_js_1.default(balance).minus(new bignumber_js_1.default(totalFeeEstimate)).toString(),
            });
            unsignedTransaction = await txBuilder.build();
            feeEstimate = totalFeeEstimate;
            transactionType = 'ton';
        }
        if (!isUnsignedSweep) {
            if (!params.userKey) {
                throw new Error('missing userKey');
            }
            if (!params.backupKey) {
                throw new Error('missing backupKey');
            }
            if (!params.walletPassphrase) {
                throw new Error('missing wallet passphrase');
            }
            // Clean up whitespace from entered values
            const userKey = params.userKey.replace(/\s/g, '');
            const backupKey = params.backupKey.replace(/\s/g, '');
            let userPrv;
            try {
                userPrv = this.bitgo.decrypt({
                    input: userKey,
                    password: params.walletPassphrase,
                });
            }
            catch (e) {
                throw new Error(`Error decrypting user keychain: ${e.message}`);
            }
            const userSigningMaterial = JSON.parse(userPrv);
            let backupPrv;
            try {
                backupPrv = this.bitgo.decrypt({
                    input: backupKey,
                    password: params.walletPassphrase,
                });
            }
            catch (e) {
                throw new Error(`Error decrypting backup keychain: ${e.message}`);
            }
            const backupSigningMaterial = JSON.parse(backupPrv);
            const signatureHex = await sdk_core_1.EDDSAMethods.getTSSSignature(userSigningMaterial, backupSigningMaterial, currPath, unsignedTransaction);
            const publicKeyObj = { pub: senderAddr };
            txBuilder.addSignature(publicKeyObj, signatureHex);
        }
        const walletCoin = this.getChain();
        const coinSpecific = { commonKeychain: bitgoKey };
        if (isUnsignedSweep) {
            const completedTransaction = await txBuilder.build();
            const serializedTx = completedTransaction.toBroadcastFormat();
            const inputs = [];
            for (const input of completedTransaction.inputs) {
                inputs.push({
                    address: input.address,
                    valueString: input.value,
                    value: new bignumber_js_1.default(input.value).toNumber(),
                });
            }
            const outputs = [];
            for (const output of completedTransaction.outputs) {
                outputs.push({
                    address: output.address,
                    valueString: output.value,
                    coinName: output.coin,
                });
            }
            const spendAmount = completedTransaction.inputs.length === 1 ? completedTransaction.inputs[0].value : 0;
            const parsedTx = { inputs: inputs, outputs: outputs, spendAmount: spendAmount, type: transactionType };
            const feeInfo = { fee: feeEstimate, feeString: feeEstimate.toString() };
            const transaction = {
                serializedTx: serializedTx,
                scanIndex: index,
                coin: walletCoin,
                signableHex: completedTransaction.signablePayload.toString('hex'),
                derivationPath: currPath,
                parsedTx: parsedTx,
                feeInfo: feeInfo,
                coinSpecific: coinSpecific,
            };
            return transaction;
        }
        const completedTransaction = await txBuilder.build();
        const serializedTx = completedTransaction.toBroadcastFormat();
        const transaction = {
            serializedTx: serializedTx,
            scanIndex: index,
        };
        return transaction;
    }
    /**
     * Creates funds sweep recovery transaction(s) without BitGo
     *
     * @param {MPCSweepRecoveryOptions} params parameters needed to combine the signatures
     * and transactions to create broadcastable transactions
     *
     * @returns {MPCTxs} array of the serialized transaction hex strings and indices
     * of the addresses being swept
     */
    async createBroadcastableSweepTransaction(params) {
        const req = params.signatureShares;
        const broadcastableTransactions = [];
        let lastScanIndex = 0;
        for (let i = 0; i < req.length; i++) {
            const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
            const transaction = req[i].txRequest.transactions[0].unsignedTx;
            if (!req[i].ovc || !req[i].ovc[0].eddsaSignature) {
                throw new Error('Missing signature(s)');
            }
            const signature = req[i].ovc[0].eddsaSignature;
            if (!transaction.signableHex) {
                throw new Error('Missing signable hex');
            }
            const messageBuffer = Buffer.from(transaction.signableHex, 'hex');
            const result = MPC.verify(messageBuffer, signature);
            if (!result) {
                throw new Error('Invalid signature');
            }
            const signatureHex = Buffer.concat([Buffer.from(signature.R, 'hex'), Buffer.from(signature.sigma, 'hex')]);
            const txBuilder = this.getBuilder().from(transaction.serializedTx);
            if (!transaction.coinSpecific?.commonKeychain) {
                throw new Error('Missing common keychain');
            }
            const commonKeychain = transaction.coinSpecific.commonKeychain;
            if (!transaction.derivationPath) {
                throw new Error('Missing derivation path');
            }
            const derivationPath = transaction.derivationPath;
            const accountId = MPC.deriveUnhardened(commonKeychain, derivationPath).slice(0, 64);
            const tonKeyPair = new keyPair_1.KeyPair({ pub: accountId });
            // add combined signature from ovc
            txBuilder.addSignature({ pub: tonKeyPair.getKeys().pub }, signatureHex);
            const signedTransaction = await txBuilder.build();
            const serializedTx = signedTransaction.toBroadcastFormat();
            broadcastableTransactions.push({
                serializedTx: serializedTx,
                scanIndex: transaction.scanIndex,
            });
            if (i === req.length - 1 && transaction.coinSpecific.lastScanIndex) {
                lastScanIndex = transaction.coinSpecific.lastScanIndex;
            }
        }
        return { transactions: broadcastableTransactions, lastScanIndex };
    }
    /** @inheritDoc */
    auditDecryptedKey({ publicKey, prv, multiSigType }) {
        if (multiSigType !== 'tss') {
            throw new Error('Unsupported multiSigType');
        }
        (0, sdk_lib_mpc_1.auditEddsaPrivateKey)(prv, publicKey ?? '');
    }
}
exports.Ton = Ton;
