// @bitgo/sdk-coin-opeth/dist/src/opethToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpethToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class OpethToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, OpethToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, OpethToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(OpethToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /**
     * Make a query to Optimism Etherscan for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response from Optimism Etherscan
     */
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].optimisticEtherscanApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].optimisticEtherscanBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    getFullName() {
        return 'Opeth Token';
    }
    supportsTss() {
        return true;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
}
exports.OpethToken = OpethToken;
OpethToken.coinNames = {
    Mainnet: 'opeth',
    Testnet: 'topeth',
};
