// @bitgo/sdk-coin-sol/dist/cjs/test/unit/messages/messageBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sinon_1 = __importDefault(require("sinon"));
const should_1 = __importDefault(require("should"));
const src_1 = require("../../../src");
const sdk_core_1 = require("@bitgo/sdk-core");
describe('Solana MessageBuilderFactory', function () {
    let sandbox;
    let factory;
    const coinConfig = { name: 'tsol' };
    // Common test data
    const signatureBase64 = Buffer.from('signature1').toString('base64');
    const testBroadcastMessage = {
        type: sdk_core_1.MessageStandardType.SIMPLE,
        payload: 'test message',
        serializedSignatures: [
            {
                publicKey: 'pubkey1',
                signature: signatureBase64,
            },
        ],
        signers: ['signer1'],
        metadata: {},
    };
    const unsupportedMessageTypes = [
        sdk_core_1.MessageStandardType.UNKNOWN,
        sdk_core_1.MessageStandardType.EIP191,
        sdk_core_1.MessageStandardType.CIP8,
        'UNSUPPORTED',
    ];
    // Helper functions
    const assertSimpleMessageBuilder = (builder) => {
        should_1.default.exist(builder);
        builder.should.be.instanceof(sdk_core_1.SimpleMessageBuilder);
    };
    const assertBuilderMessageProperties = async (builder, expectedPayload) => {
        const message = await builder.build();
        message.getType().should.equal(sdk_core_1.MessageStandardType.SIMPLE);
        message.getPayload().should.equal(expectedPayload);
        message.getSignatures().should.have.length(1);
        message.getSigners().should.have.length(1);
        message.getSigners()[0].should.equal('signer1');
        message.getMetadata().should.be.an.Object();
        return message;
    };
    const assertSignatureProperties = (message) => {
        const signature = message.getSignatures()[0];
        signature.should.have.properties(['publicKey', 'signature']);
        signature.publicKey.pub.should.equal('pubkey1');
        signature.signature.toString('base64').should.equal(signatureBase64);
    };
    const assertBroadcastFormatProperties = async (message, expectedSerializedSignatures) => {
        const rebroadcastMessage = await message.toBroadcastFormat();
        rebroadcastMessage.should.have.properties(['type', 'payload', 'serializedSignatures', 'signers', 'metadata']);
        rebroadcastMessage.type.should.equal(sdk_core_1.MessageStandardType.SIMPLE);
        rebroadcastMessage.payload.should.equal('test message');
        rebroadcastMessage.serializedSignatures?.should.deepEqual(expectedSerializedSignatures);
        rebroadcastMessage.signers?.should.deepEqual(['signer1']);
    };
    beforeEach(function () {
        sandbox = sinon_1.default.createSandbox();
        factory = new src_1.MessageBuilderFactory(coinConfig);
    });
    afterEach(function () {
        sandbox.restore();
    });
    describe('getMessageBuilder', function () {
        it('should return SimpleMessageBuilder for SIMPLE type', function () {
            const builder = factory.getMessageBuilder(sdk_core_1.MessageStandardType.SIMPLE);
            assertSimpleMessageBuilder(builder);
        });
        it('should throw error for unsupported message type', function () {
            unsupportedMessageTypes.forEach((type) => {
                should_1.default.throws(() => factory.getMessageBuilder(type), new RegExp(`Invalid message standard ${type}`));
            });
        });
    });
    describe('fromBroadcastFormat', function () {
        it('should get the correct builder type from broadcastable message', async function () {
            const builder = factory.fromBroadcastFormat(testBroadcastMessage);
            assertSimpleMessageBuilder(builder);
            const message = await assertBuilderMessageProperties(builder, 'test message');
            assertSignatureProperties(message);
            await assertBroadcastFormatProperties(message, testBroadcastMessage.serializedSignatures);
        });
        it('should throw for unsupported message type in broadcastable message', function () {
            const broadcastMessage = {
                ...testBroadcastMessage,
                type: sdk_core_1.MessageStandardType.EIP191,
            };
            should_1.default.throws(() => factory.fromBroadcastFormat(broadcastMessage), new RegExp(`Invalid message standard ${sdk_core_1.MessageStandardType.EIP191}`));
        });
    });
    describe('fromBroadcastString', function () {
        it('should parse broadcastable string and return correct builder type', async function () {
            const broadcastString = JSON.stringify(testBroadcastMessage);
            const broadcastHex = Buffer.from(broadcastString).toString('hex');
            const builder = factory.fromBroadcastString(broadcastHex);
            assertSimpleMessageBuilder(builder);
            const message = await assertBuilderMessageProperties(builder, 'test message');
            assertSignatureProperties(message);
            await assertBroadcastFormatProperties(message, testBroadcastMessage.serializedSignatures);
        });
        it('should throw for invalid JSON string', function () {
            try {
                factory.fromBroadcastString('abcdefg'); // Invalid hex string
                fail('Expected error not thrown');
            }
            catch (error) {
                error.should.be.instanceof(SyntaxError);
            }
        });
    });
});
