// @bitgo/sdk-coin-canton/dist/src/lib/walletInitialization/walletInitTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletInitTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const constant_1 = require("../constant");
class WalletInitTransaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
    }
    get preparedParty() {
        return this._preparedParty;
    }
    set preparedParty(transaction) {
        this._preparedParty = transaction;
        this._id = transaction.multiHash;
        this._type = sdk_core_1.TransactionType.WalletInitialization;
    }
    canSign(key) {
        return false;
    }
    set signatures(signature) {
        this._signatures.push(signature);
    }
    toBroadcastFormat() {
        if (!this._preparedParty) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        const multiHashSignatures = [];
        if (this.signature.length > 0) {
            this.signature.map((signature) => {
                const multiHashSignature = {
                    format: constant_1.SIGNATURE_FORMAT,
                    signature: signature,
                    signedBy: this._preparedParty.publicKeyFingerprint,
                    signingAlgorithmSpec: constant_1.SIGNATURE_ALGORITHM_SPEC,
                };
                multiHashSignatures.push(multiHashSignature);
            });
        }
        const walletInitBroadcastData = {
            preparedParty: this._preparedParty,
            onboardingTransactions: this._preparedParty.topologyTransactions.map((txn) => {
                const onboardingTransaction = {
                    transaction: txn,
                };
                return onboardingTransaction;
            }),
            multiHashSignatures: multiHashSignatures,
        };
        return Buffer.from(JSON.stringify(walletInitBroadcastData)).toString('base64');
    }
    toJson() {
        if (!this._preparedParty) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        const result = {
            id: this.id,
            type: this._type,
            preparedParty: this._preparedParty,
        };
        return result;
    }
    get signablePayload() {
        if (!this._preparedParty) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction data');
        }
        return Buffer.from(this._preparedParty.multiHash, 'base64');
    }
    fromRawTransaction(rawTx) {
        try {
            const decoded = JSON.parse(Buffer.from(rawTx, 'base64').toString('utf8'));
            this._preparedParty = decoded.preparedParty;
            this._type = sdk_core_1.TransactionType.WalletInitialization;
            this._id = decoded.preparedParty.multiHash;
            if (decoded.multiHashSignatures.length > 0) {
                decoded.multiHashSignatures.map((multiHashSignature) => {
                    this.signatures = multiHashSignature.signature;
                });
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError('Unable to parse raw transaction data');
        }
    }
    explainTransaction() {
        const displayOrder = ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'];
        return {
            id: this.id,
            displayOrder,
            outputs: [],
            outputAmount: '0',
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: '0' },
            type: this.type,
        };
    }
}
exports.WalletInitTransaction = WalletInitTransaction;
