// @bitgo/sdk-coin-canton/dist/src/cantonToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CantonToken = void 0;
const statics_1 = require("@bitgo/statics");
const canton_1 = require("./canton");
class CantonToken extends canton_1.Canton {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === statics_1.NetworkType.MAINNET ? statics_1.coins.get('canton') : statics_1.coins.get('tcanton');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new CantonToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfig = [...statics_1.tokens.bitcoin.canton.tokens, ...statics_1.tokens.testnet.canton.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfig) {
            const tokenConstructor = CantonToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get baseUrl() {
        return this.tokenConfig.baseUrl;
    }
    get contractAddress() {
        return this.tokenConfig.contractAddress;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Canton Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.CantonToken = CantonToken;
