// @bitgo/sdk-coin-canton/dist/src/canton.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Canton = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
const keyPair_1 = require("./lib/keyPair");
const utils_1 = __importDefault(require("./lib/utils"));
class Canton extends sdk_core_1.BaseCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Canton(bitgo, staticsCoin);
    }
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    /** @inheritDoc */
    getBaseFactor() {
        return 1e10;
    }
    /** @inheritDoc */
    getChain() {
        return 'canton';
    }
    /** @inheritDoc */
    getFamily() {
        return 'canton';
    }
    /** @inheritDoc */
    getFullName() {
        return 'Canton';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return sdk_core_1.multisigTypes.tss;
    }
    /** inherited doc */
    requiresWalletInitializationTransaction() {
        return true;
    }
    getMPCAlgorithm() {
        return 'eddsa';
    }
    /** @inheritDoc */
    async verifyTransaction(params) {
        const coinConfig = statics_1.coins.get(this.getChain());
        const { txPrebuild: txPrebuild, txParams } = params;
        const rawTx = txPrebuild.txHex;
        if (!rawTx) {
            throw new Error('missing required tx prebuild property txHex');
        }
        const txBuilder = new lib_1.TransactionBuilderFactory(coinConfig).from(rawTx);
        const transaction = txBuilder.transaction;
        const explainedTx = transaction.explainTransaction();
        switch (transaction.type) {
            case sdk_core_1.TransactionType.WalletInitialization:
            case sdk_core_1.TransactionType.TransferAccept:
            case sdk_core_1.TransactionType.TransferReject:
            case sdk_core_1.TransactionType.TransferAcknowledge:
            case sdk_core_1.TransactionType.OneStepPreApproval:
            case sdk_core_1.TransactionType.TransferOfferWithdrawn:
                // There is no input for these type of transactions, so always return true.
                return true;
            case sdk_core_1.TransactionType.Send:
                if (txParams.recipients !== undefined) {
                    const filteredRecipients = txParams.recipients?.map((recipient) => {
                        const { address, amount, tokenName } = recipient;
                        const [addressPart, memoId] = address.split('?memoId=');
                        return {
                            address: addressPart,
                            amount,
                            ...(memoId && { memo: memoId }),
                            ...(tokenName && { tokenName }),
                        };
                    });
                    const filteredOutputs = explainedTx.outputs?.map((output) => {
                        const { address, amount, tokenName, memo } = output;
                        return {
                            address,
                            amount,
                            ...(memo && { memo }),
                            ...(tokenName && { tokenName }),
                        };
                    });
                    if (JSON.stringify(filteredRecipients) !== JSON.stringify(filteredOutputs)) {
                        throw new Error('Tx outputs do not match with expected txParams recipients');
                    }
                }
                return true;
            default: {
                throw new Error(`unknown transaction type, ${transaction.type}`);
            }
        }
    }
    /** @inheritDoc */
    async isWalletAddress(params) {
        // TODO: refactor this and use the `verifyEddsaMemoBasedWalletAddress` once published from sdk-core
        // https://bitgoinc.atlassian.net/browse/COIN-6347
        const { keychains, address: newAddress, index } = params;
        const [addressPart, memoId] = newAddress.split('?memoId=');
        if (!this.isValidAddress(addressPart)) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${newAddress}`);
        }
        if (memoId && memoId !== `${index}`) {
            throw new sdk_core_1.InvalidAddressError(`invalid memoId index: ${memoId}`);
        }
        const commonKeychain = (0, sdk_core_1.extractCommonKeychain)(keychains);
        const MPC = await sdk_core_1.EDDSAMethods.getInitializedMpcInstance();
        const derivationPath = 'm/0';
        const derivedPublicKey = MPC.deriveUnhardened(commonKeychain, derivationPath).slice(0, 64);
        const publicKeyBase64 = Buffer.from(derivedPublicKey, 'hex').toString('base64');
        const rootAddressFingerprint = utils_1.default.getAddressFromPublicKey(publicKeyBase64);
        const rootAddress = `${rootAddressFingerprint.slice(0, 5)}::${rootAddressFingerprint}`;
        return addressPart === rootAddress;
    }
    /** @inheritDoc */
    parseTransaction(params) {
        throw new Error('Method not implemented.');
    }
    /** @inheritDoc */
    generateKeyPair(seed) {
        const keyPair = seed ? new keyPair_1.KeyPair({ seed }) : new keyPair_1.KeyPair();
        const keys = keyPair.getKeys();
        if (!keys.prv) {
            throw new Error('Missing prv in key generation.');
        }
        return {
            pub: keys.pub,
            prv: keys.prv,
        };
    }
    /** @inheritDoc */
    explainTransaction(params) {
        const factory = this.getBuilder();
        let rebuiltTransaction;
        const txRaw = params.txHex;
        try {
            const txBuilder = factory.from(txRaw);
            rebuiltTransaction = txBuilder.transaction;
        }
        catch (e) {
            throw new Error('Invalid transaction');
        }
        return rebuiltTransaction.explainTransaction();
    }
    /** @inheritDoc */
    isValidPub(pub) {
        return utils_1.default.isValidPublicKey(pub);
    }
    /** @inheritDoc */
    isValidAddress(address) {
        // canton addresses are of the form, partyHint::fingerprint
        // where partyHint is of length 5 and fingerprint is 68 characters long
        return utils_1.default.isValidAddress(address);
    }
    /** @inheritDoc */
    getTokenEnablementConfig() {
        return {
            requiresTokenEnablement: true,
            supportsMultipleTokenEnablements: false,
        };
    }
    getAddressFromPublicKey(publicKeyHex) {
        const publicKeyBase64 = Buffer.from(publicKeyHex, 'hex').toString('base64');
        return utils_1.default.getAddressFromPublicKey(publicKeyBase64);
    }
    /** @inheritDoc */
    signTransaction(params) {
        throw new Error('Method not implemented.');
    }
    /** @inheritDoc */
    auditDecryptedKey({ multiSigType, prv, publicKey }) {
        if (multiSigType !== sdk_core_1.multisigTypes.tss) {
            throw new Error('Unsupported multiSigType');
        }
        (0, sdk_lib_mpc_1.auditEddsaPrivateKey)(prv, publicKey ?? '');
    }
    /** @inheritDoc */
    setCoinSpecificFieldsInIntent(intent, params) {
        if (params.txRequestId) {
            intent.txRequestId = params.txRequestId;
        }
        if (params.transferOfferId) {
            intent.transferOfferId = params.transferOfferId;
        }
    }
}
exports.Canton = Canton;
