// @bitgo/sdk-coin-polyx/dist/src/polyxToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolyxToken = void 0;
const statics_1 = require("@bitgo/statics");
const polyx_1 = require("./polyx");
class PolyxToken extends polyx_1.Polyx {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('polyx') : statics_1.coins.get('tpolyx');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new PolyxToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.polyx.tokens, ...statics_1.tokens.testnet.polyx.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            const tokenConstructor = PolyxToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get ticker() {
        return this.tokenConfig.ticker;
    }
    get assetId() {
        return this.tokenConfig.assetId;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return this._staticsCoin.fullName;
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.PolyxToken = PolyxToken;
