// @bitgo/utxo-core/dist/cjs/test/descriptor/fromFixedScriptWallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("node:assert/strict"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const fromFixedScriptWallet_1 = require("../../src/descriptor/fromFixedScriptWallet");
function getRootWalletKeys(derivationPrefixes) {
    // This is a fixed script wallet, so we use a fixed key triple.
    // In practice, this would be derived from the wallet's root keys.
    return new utxolib.bitgo.RootWalletKeys(utxolib.testutil.getKeyTriple('fixedScript'), derivationPrefixes);
}
const customPrefixes = [
    undefined,
    ['1/2', '1/2', '1/2'], // Custom prefixes for testing
    ['1/2', '3/4', '5/6'], // Different custom prefixes
];
const scriptTypes = utxolib.bitgo.outputScripts.scriptTypes2Of3;
const scope = ['external', 'internal'];
const index = [0, 1, 2];
/** Get the expected max weight to satisfy the descriptor */
function getExpectedMaxWeightToSatisfy(scriptType) {
    switch (scriptType) {
        case 'p2sh':
            return 256;
        case 'p2shP2wsh':
            return 99;
        case 'p2wsh':
            return 64;
        default:
            throw new Error('unexpected script type');
    }
}
/** Compute the total size of the input, including overhead */
function getTotalInputSize(vSize) {
    const sizeOpPushData1 = 1;
    const sizeOpPushData2 = 2;
    return 32 /* txid */ + 4 /* vout */ + 4 /* nSequence */ + (vSize < 255 ? sizeOpPushData1 : sizeOpPushData2) + vSize;
}
/** Get the full expected vSize of the input including overhead */
function getExpectedVSize(scriptType) {
    // https://github.com/BitGo/BitGoJS/blob/master/modules/unspents/docs/input-costs.md
    switch (scriptType) {
        case 'p2sh':
            return 298;
        case 'p2shP2wsh':
            return 140;
        case 'p2wsh':
            return 105;
        default:
            throw new Error('unexpected script type');
    }
}
function runTestGetDescriptorWithScriptType(customPrefix, scriptType, index, scope) {
    describe(`customPrefix=${customPrefix}, scriptType=${scriptType}, index=${index}, scope=${scope}`, function () {
        const rootWalletKeys = getRootWalletKeys(customPrefix);
        const chainCode = scope === 'external'
            ? utxolib.bitgo.getExternalChainCode(scriptType)
            : utxolib.bitgo.getInternalChainCode(scriptType);
        const derivedKeys = rootWalletKeys.deriveForChainAndIndex(chainCode, index);
        const scriptUtxolib = utxolib.bitgo.outputScripts.createOutputScript2of3(derivedKeys.publicKeys, scriptType).scriptPubKey;
        let descriptor;
        before(function () {
            descriptor = (0, fromFixedScriptWallet_1.getDescriptorForScriptType)(rootWalletKeys, scriptType, scope);
        });
        it('address should match descriptor', function () {
            const scriptFromDescriptor = Buffer.from(descriptor.atDerivationIndex(index).scriptPubkey());
            assert.deepStrictEqual(scriptUtxolib.toString('hex'), scriptFromDescriptor.toString('hex'));
        });
        it('should have expected weights', function () {
            assert.ok(Number.isInteger(descriptor.maxWeightToSatisfy()));
            const vSize = Math.ceil(descriptor.maxWeightToSatisfy() / 4);
            assert.equal(vSize, getExpectedMaxWeightToSatisfy(scriptType));
            assert.equal(getTotalInputSize(vSize), getExpectedVSize(scriptType));
        });
    });
}
function runTestGetDescriptorMap(customPrefix) {
    describe(`getNamedDescriptorsForRootWalletKeys with customPrefix=${customPrefix}`, function () {
        const rootWalletKeys = getRootWalletKeys(customPrefix);
        const descriptorMap = (0, fromFixedScriptWallet_1.getNamedDescriptorsForRootWalletKeys)(rootWalletKeys);
        it('should return a map with the correct number of entries', function () {
            assert.equal(descriptorMap.size, scriptTypes.length * scope.length);
        });
        scriptTypes.forEach((scriptType) => {
            scope.forEach((s) => {
                const key = `${scriptType}/${s}`;
                it(`should have a correct descriptor for ${key}`, function () {
                    const descriptorFromMap = descriptorMap.get(key);
                    if (scriptType === 'p2tr' || scriptType === 'p2trMusig2') {
                        assert.equal(descriptorFromMap, null, `Descriptor for ${key} should be null`);
                        return;
                    }
                    assert.ok(descriptorFromMap, `Descriptor for ${key} should exist`);
                    const expectedDescriptor = (0, fromFixedScriptWallet_1.getDescriptorForScriptType)(rootWalletKeys, scriptType, s);
                    assert.equal(descriptorFromMap.toString(), expectedDescriptor.toString());
                });
            });
        });
    });
}
customPrefixes.forEach((customPrefix) => {
    scriptTypes.forEach((scriptType) => {
        index.forEach((index) => {
            scope.forEach((scope) => {
                if (scriptType !== 'p2tr' && scriptType !== 'p2trMusig2') {
                    runTestGetDescriptorWithScriptType(customPrefix, scriptType, index, scope);
                }
            });
        });
    });
    runTestGetDescriptorMap(customPrefix);
});
