// @bitgo/utxo-core/dist/cjs/test/descriptor/Output.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const Output_1 = require("../../src/descriptor/Output");
const descriptor_1 = require("../../src/testutil/descriptor");
const descriptor_2 = require("../../src/descriptor");
describe('decscriptor.Output', function () {
    const descriptor = (0, descriptor_1.getDescriptor)('Wsh2Of3');
    it('isInternalOutput correctly identifies internal outputs', function () {
        const internalOutput = { value: 1n, descriptor };
        const externalOutput = { value: 1n };
        assert_1.default.strictEqual((0, Output_1.isInternalOutput)(internalOutput), true);
        assert_1.default.strictEqual((0, Output_1.isInternalOutput)(externalOutput), false);
    });
    it('isExternalOutput correctly identifies external outputs', function () {
        const internalOutput = { value: 1n, descriptor };
        const externalOutput = { value: 1n };
        assert_1.default.strictEqual((0, Output_1.isExternalOutput)(internalOutput), false);
        assert_1.default.strictEqual((0, Output_1.isExternalOutput)(externalOutput), true);
    });
    it('toDerivedDescriptorWalletOutput returns expected values', function () {
        const derivable = descriptor;
        const definite = derivable.atDerivationIndex(0);
        for (const descriptor of [derivable, definite]) {
            const descriptorIndex = descriptor === derivable ? 0 : undefined;
            const descriptorMap = new Map([['desc', descriptor]]);
            const descriptorWalletOutput = {
                hash: Buffer.alloc(32).toString('hex'),
                index: 0,
                witnessUtxo: {
                    script: (0, descriptor_2.createScriptPubKeyFromDescriptor)(descriptor, descriptorIndex),
                    value: 1n,
                },
                descriptorName: 'desc',
                descriptorIndex,
            };
            assert_1.default.strictEqual((0, Output_1.toDerivedDescriptorWalletOutput)(descriptorWalletOutput, descriptorMap).descriptor.toString(), definite.toString());
        }
    });
});
