// @bitgo/utxo-core/dist/esm/testutil/descriptor/mock.utils.js
import * as utxolib from '@bitgo/utxo-lib';
import { createPsbt, createScriptPubKeyFromDescriptor, } from '../../descriptor';
import { getDefaultXPubs, getDescriptor } from './descriptors';
function mockOutputId(id) {
    const hash = id?.hash ?? Buffer.alloc(32, 1).toString('hex');
    const vout = id?.vout ?? 0;
    return { hash, vout };
}
export function mockDerivedDescriptorWalletOutput(descriptor, outputParams = {}) {
    const { value = BigInt(1e6) } = outputParams;
    const { hash, vout } = mockOutputId(outputParams.id);
    return {
        hash,
        index: vout,
        witnessUtxo: {
            script: createScriptPubKeyFromDescriptor(descriptor, undefined),
            value,
        },
        descriptor,
        selectTapLeafScript: outputParams.selectTapLeafScript,
        sequence: outputParams.sequence,
    };
}
function tryDeriveAtIndex(descriptor, index) {
    return descriptor.hasWildcard() ? descriptor.atDerivationIndex(index) : descriptor;
}
export function mockPsbt(inputs, outputs, params = {}) {
    return createPsbt({ ...params, network: params.network ?? utxolib.networks.bitcoin }, inputs.map((i) => mockDerivedDescriptorWalletOutput(tryDeriveAtIndex(i.descriptor, i.index), i)), outputs.map((o) => {
        const derivedDescriptor = tryDeriveAtIndex(o.descriptor, o.index);
        return {
            script: createScriptPubKeyFromDescriptor(derivedDescriptor, undefined),
            value: o.value,
            descriptor: o.external ? undefined : derivedDescriptor,
        };
    }));
}
export function mockPsbtDefault({ descriptorSelf = getDescriptor('Wsh2Of3', getDefaultXPubs('a')), descriptorOther = getDescriptor('Wsh2Of3', getDefaultXPubs('b')), params = {}, } = {}) {
    return mockPsbt([
        { descriptor: descriptorSelf, index: 0 },
        { descriptor: descriptorSelf, index: 1, id: { vout: 1 } },
    ], [
        { descriptor: descriptorOther, index: 0, value: BigInt(4e5), external: true },
        { descriptor: descriptorSelf, index: 0, value: BigInt(4e5) },
    ], params);
}
export function mockPsbtDefaultWithDescriptorTemplate(t, params = {}) {
    return mockPsbtDefault({
        descriptorSelf: getDescriptor(t, getDefaultXPubs('a')),
        descriptorOther: getDescriptor(t, getDefaultXPubs('b')),
        params,
    });
}
