// @bitgo/sdk-core/dist/src/bitgo/utils/mpcUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MpcUtils = void 0;
/**
 * @prettier
 */
const assert_1 = __importDefault(require("assert"));
const openpgp_1 = require("openpgp");
const opengpgUtils_1 = require("./opengpgUtils");
const bitgoPubKeys_1 = require("../tss/bitgoPubKeys");
const statics_1 = require("@bitgo/statics");
class MpcUtils {
    constructor(bitgo, baseCoin) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
    }
    async decryptPrivateShare(privateShare, userGpgKey) {
        const privateShareMessage = await (0, openpgp_1.readMessage)({
            armoredMessage: privateShare,
        });
        const userGpgPrivateKey = await (0, openpgp_1.readPrivateKey)({ armoredKey: userGpgKey.privateKey });
        const decryptedPrivateShare = (await (0, openpgp_1.decrypt)({
            message: privateShareMessage,
            decryptionKeys: [userGpgPrivateKey],
            format: 'utf8',
        })).data;
        return decryptedPrivateShare;
    }
    async createBitgoKeychainInWP(userGpgKey, backupGpgKey, userKeyShare, backupKeyShare, keyType, enterprise) {
        const bitgoKey = (await (0, opengpgUtils_1.getBitgoGpgPubKey)(this.bitgo)).mpcV1;
        if ((0, bitgoPubKeys_1.envRequiresBitgoPubGpgKeyConfig)(this.bitgo.getEnv())) {
            // Ensure the public key is one of the expected BitGo public keys when in test or prod.
            (0, assert_1.default)((0, bitgoPubKeys_1.isBitgoMpcPubKey)(bitgoKey.armor(), 'mpcv1'), 'Invalid BitGo GPG public key');
        }
        const encUserToBitGoMessage = await (0, opengpgUtils_1.encryptText)(userKeyShare.privateShare, bitgoKey);
        const encBackupToBitGoMessage = await (0, opengpgUtils_1.encryptText)(backupKeyShare.privateShare, bitgoKey);
        const createBitGoMPCParams = {
            keyType,
            source: 'bitgo',
            keyShares: [
                {
                    from: 'user',
                    to: 'bitgo',
                    publicShare: userKeyShare.publicShare,
                    privateShare: encUserToBitGoMessage,
                    privateShareProof: userKeyShare.privateShareProof,
                    vssProof: userKeyShare.vssProof,
                },
                {
                    from: 'backup',
                    to: 'bitgo',
                    publicShare: backupKeyShare.publicShare,
                    privateShare: encBackupToBitGoMessage,
                    privateShareProof: backupKeyShare.privateShareProof,
                    vssProof: backupKeyShare.vssProof,
                },
            ],
            userGPGPublicKey: userGpgKey.publicKey,
            backupGPGPublicKey: backupGpgKey.publicKey,
            enterprise: enterprise,
        };
        return await this.baseCoin.keychains().add(createBitGoMPCParams);
    }
    /**
     * This function would be responsible for populating intents
     * based on the type of coin / sig scheme the coin uses
     * @param {IBaseCoin} baseCoin
     * @param {PrebuildTransactionWithIntentOptions} params
     * @returns {Record<string, unknown>}
     */
    populateIntent(baseCoin, params) {
        const chain = this.baseCoin.getChain();
        if (params.intentType === 'customTx' && baseCoin.getFamily() === 'sol') {
            const hasSolInstructions = params.solInstructions && params.solInstructions.length > 0;
            const hasVersionedData = params.solVersionedTransactionData &&
                params.solVersionedTransactionData.versionedInstructions &&
                params.solVersionedTransactionData.versionedInstructions.length > 0;
            (0, assert_1.default)(hasSolInstructions || hasVersionedData, `'solInstructions' or 'solVersionedTransactionData' is required for customTx intent`);
        }
        if (params.intentType === 'customTx' && baseCoin.getFamily() === 'apt') {
            (0, assert_1.default)(params.aptosCustomTransactionParams, `'aptosCustomTransactionParams' is a required parameter for customTx intent`);
        }
        if (['transferAccept', 'transferReject'].includes(params.intentType) && baseCoin.getFamily() === 'canton') {
            (0, assert_1.default)(params.txRequestId, `'txRequestId' is required parameter for ${params.intentType} intent`);
        }
        if (params.intentType === 'transferOfferWithdrawn' && baseCoin.getFamily() === 'canton') {
            (0, assert_1.default)(params.transferOfferId, `'transferOfferId' is required parameter for ${params.intentType} intent`);
        }
        if (![
            'acceleration',
            'fillNonce',
            'transferToken',
            'tokenApproval',
            'customTx',
            'transferAccept',
            'transferReject',
            'transferOfferWithdrawn',
            'bridgeFunds',
        ].includes(params.intentType)) {
            (0, assert_1.default)(params.recipients, `'recipients' is a required parameter for ${params.intentType} intent`);
        }
        const intentRecipients = params.recipients?.map((recipient) => {
            const formattedRecipient = {
                address: { address: recipient.address },
                amount: { value: `${recipient.amount}`, symbol: recipient.tokenName ? recipient.tokenName : chain },
            };
            if (recipient.data) {
                formattedRecipient.data = recipient.data;
            }
            const { tokenData } = recipient;
            if (tokenData && (tokenData.tokenContractAddress || tokenData.tokenName)) {
                // token related recipient data gets validated in WP
                if (!(tokenData.tokenType && tokenData.tokenQuantity)) {
                    throw new Error('token type and quantity is required to request a transaction with intent to transfer a token');
                }
                tokenData.tokenName = this.getTokenName(baseCoin, tokenData);
                if (tokenData.tokenName) {
                    formattedRecipient.amount.symbol = tokenData.tokenName;
                }
                formattedRecipient.tokenData = tokenData;
            }
            return formattedRecipient;
        });
        const baseIntent = {
            intentType: params.intentType,
            sequenceId: params.sequenceId,
            comment: params.comment,
            nonce: params.nonce,
            recipients: intentRecipients,
            tokenName: params.tokenName,
            isTestTransaction: params.isTestTransaction,
        };
        if (baseCoin.isEVM() && baseCoin.supportsTss()) {
            switch (params.intentType) {
                case 'payment':
                case 'transferToken':
                case 'fillNonce':
                    return {
                        ...baseIntent,
                        selfSend: params.selfSend,
                        feeOptions: params.feeOptions,
                        hopParams: params.hopParams,
                        isTss: params.isTss,
                        nonce: params.nonce,
                        custodianTransactionId: params.custodianTransactionId,
                        receiveAddress: params.receiveAddress,
                    };
                case 'acceleration':
                    return {
                        ...baseIntent,
                        txid: params.lowFeeTxid,
                        receiveAddress: params.receiveAddress,
                        feeOptions: params.feeOptions,
                    };
                case 'tokenApproval':
                    return {
                        ...baseIntent,
                        tokenName: params.tokenName,
                        feeOptions: params.feeOptions,
                    };
                case 'bridgeFunds':
                    return {
                        ...baseIntent,
                        amount: params.amount,
                        feeOptions: params.feeOptions,
                    };
                default:
                    throw new Error(`Unsupported intent type ${params.intentType}`);
            }
        }
        this.baseCoin.setCoinSpecificFieldsInIntent(baseIntent, params);
        if (params.feeOptions !== undefined) {
            return {
                ...baseIntent,
                memo: params.memo?.value,
                token: params.tokenName,
                enableTokens: params.enableTokens,
                feeOptions: params.feeOptions,
            };
        }
        return {
            ...baseIntent,
            memo: params.memo?.value,
            token: params.tokenName,
            enableTokens: params.enableTokens,
        };
    }
    getTokenName(baseCoin, tokenData) {
        if (tokenData.tokenName) {
            return tokenData.tokenName;
        }
        const networkPrefix = baseCoin.getConfig().network.type === statics_1.NetworkType.TESTNET ? 't' : '';
        const tokenStaticsKey = `${networkPrefix}${baseCoin.getFamily()}:${tokenData.tokenContractAddress}`;
        if (statics_1.coins.has(tokenStaticsKey)) {
            const tokenStatics = statics_1.coins.get(tokenStaticsKey);
            tokenData.tokenName = tokenStatics.name;
        }
        return tokenData.tokenName;
    }
}
exports.MpcUtils = MpcUtils;
