// micro-eth-signer/lib/contracts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CONTRACTS = exports.tokenFromSymbol = exports.COMMON_TOKENS = exports.WETH_CONTRACT = exports.KYBER_NETWORK_PROXY_CONTRACT = exports.UNISWAP_V3_ROUTER_CONTRACT = exports.UNISWAP_V2_ROUTER_CONTRACT = exports.WETH = exports.KYBER_NETWORK_PROXY = exports.UNISWAP_V3_ROUTER = exports.UNISWAP_V2_ROUTER = exports.ERC721 = exports.ERC20 = void 0;
var erc20_js_1 = require("./erc20.js");
Object.defineProperty(exports, "ERC20", { enumerable: true, get: function () { return erc20_js_1.default; } });
var erc721_js_1 = require("./erc721.js");
Object.defineProperty(exports, "ERC721", { enumerable: true, get: function () { return erc721_js_1.default; } });
const uniswap_v2_js_1 = require("./uniswap-v2.js");
Object.defineProperty(exports, "UNISWAP_V2_ROUTER", { enumerable: true, get: function () { return uniswap_v2_js_1.default; } });
const uniswap_v3_js_1 = require("./uniswap-v3.js");
Object.defineProperty(exports, "UNISWAP_V3_ROUTER", { enumerable: true, get: function () { return uniswap_v3_js_1.default; } });
const kyber_js_1 = require("./kyber.js");
Object.defineProperty(exports, "KYBER_NETWORK_PROXY", { enumerable: true, get: function () { return kyber_js_1.default; } });
const weth_js_1 = require("./weth.js");
Object.defineProperty(exports, "WETH", { enumerable: true, get: function () { return weth_js_1.default; } });
exports.UNISWAP_V2_ROUTER_CONTRACT = '0x7a250d5630b4cf539739df2c5dacb4c659f2488d';
exports.UNISWAP_V3_ROUTER_CONTRACT = '0xe592427a0aece92de3edee1f18e0157c05861564';
exports.KYBER_NETWORK_PROXY_CONTRACT = '0x9aab3f75489902f3a48495025729a0af77d4b11e';
exports.WETH_CONTRACT = '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2';
exports.COMMON_TOKENS = {};
const ERC20 = [
    ['UNI', '0x1f9840a85d5af5bf1d1762f925bdaddc4201f984'],
    ['BAT', '0x0d8775f648430679a709e98d2b0cb6250d2887ef'],
    // Required for Uniswap multi-hop routing
    ['USDT', '0xdac17f958d2ee523a2206206994597c13d831ec7', 6, 1],
    ['USDC', '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', 6, 1],
    ['WETH', '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'],
    ['WBTC', '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599', 8],
    ['DAI', '0x6b175474e89094c44da98b954eedeac495271d0f', 18, 1],
    ['COMP', '0xc00e94cb662c3520282e6f5717214004a7f26888'],
    ['MKR', '0x9f8f72aa9304c8b593d555f12ef6589cc3a579a2'],
    ['AMPL', '0xd46ba6d942050d489dbd938a2c909a5d5039a161', 9],
];
for (let item of ERC20) {
    const [symbol, addr, decimals, price] = item;
    exports.COMMON_TOKENS[addr] = { abi: 'ERC20', symbol, decimals: decimals || 18, price };
}
function tokenFromSymbol(symbol) {
    for (let c in exports.COMMON_TOKENS) {
        if (exports.COMMON_TOKENS[c].symbol === symbol)
            return Object.assign({ contract: c }, exports.COMMON_TOKENS[c]);
    }
    throw new Error('unknown token');
}
exports.tokenFromSymbol = tokenFromSymbol;
exports.DEFAULT_CONTRACTS = {
    [exports.UNISWAP_V2_ROUTER_CONTRACT]: { abi: uniswap_v2_js_1.default, name: 'UNISWAP V2 ROUTER' },
    [exports.KYBER_NETWORK_PROXY_CONTRACT]: { abi: kyber_js_1.default, name: 'KYBER NETWORK PROXY' },
    [exports.UNISWAP_V3_ROUTER_CONTRACT]: { abi: uniswap_v3_js_1.default, name: 'UNISWAP V3 ROUTER' },
    ...exports.COMMON_TOKENS,
    [exports.WETH_CONTRACT]: { abi: weth_js_1.default, name: 'WETH Token', decimals: 18, symbol: 'WETH' },
};
