// @bitgo/sdk-coin-vet/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.feeEstimateData = exports.COEF_DIVISOR = exports.GAS_UNIT_PRICE = exports.GAS_PRICE_COEF = exports.EXPIRATION = exports.AVG_GAS_UNITS = exports.VALIDATOR_REGISTRATION_STAKER_CONTRACT_ADDRESS_MAINNET = exports.VALIDATOR_REGISTRATION_STAKER_CONTRACT_ADDRESS_TESTNET = exports.STARGATE_CONTRACT_ADDRESS = exports.STARGATE_CONTRACT_ADDRESS_TESTNET = exports.STARGATE_NFT_ADDRESS_TESTNET = exports.STARGATE_NFT_ADDRESS = exports.CLAIM_STAKING_REWARDS_METHOD_ID = exports.CLAIM_BASE_REWARDS_METHOD_ID = exports.TRANSFER_NFT_METHOD_ID = exports.BURN_NFT_METHOD_ID = exports.EXIT_DELEGATION_METHOD_ID = exports.INCREASE_STAKE_METHOD_ID = exports.ADD_VALIDATION_METHOD_ID = exports.DELEGATE_CLAUSE_METHOD_ID = exports.STAKE_CLAUSE_METHOD_ID = exports.STAKING_METHOD_ID = exports.TRANSFER_TOKEN_METHOD_ID = exports.ZERO_VALUE_AMOUNT = exports.VET_BLOCK_ID_LENGTH = exports.VET_ADDRESS_LENGTH = exports.VET_TRANSACTION_ID_LENGTH = void 0;
exports.VET_TRANSACTION_ID_LENGTH = 64;
exports.VET_ADDRESS_LENGTH = 40;
exports.VET_BLOCK_ID_LENGTH = 64;
exports.ZERO_VALUE_AMOUNT = '0';
exports.TRANSFER_TOKEN_METHOD_ID = '0xa9059cbb';
exports.STAKING_METHOD_ID = '0xd8da3bbf';
exports.STAKE_CLAUSE_METHOD_ID = '0x604f2177';
exports.DELEGATE_CLAUSE_METHOD_ID = '0x08bbb824';
exports.ADD_VALIDATION_METHOD_ID = '0xc3c4b138';
exports.INCREASE_STAKE_METHOD_ID = '0x43b0de9a';
exports.EXIT_DELEGATION_METHOD_ID = '0x69e79b7d';
exports.BURN_NFT_METHOD_ID = '0x2e17de78';
exports.TRANSFER_NFT_METHOD_ID = '0x23b872dd';
exports.CLAIM_BASE_REWARDS_METHOD_ID = '0x858d50e8'; // claimVetGeneratedVtho(uint256)
exports.CLAIM_STAKING_REWARDS_METHOD_ID = '0x0962ef79'; // claimRewards(uint256)
exports.STARGATE_NFT_ADDRESS = '0x1856c533ac2d94340aaa8544d35a5c1d4a21dee7';
exports.STARGATE_NFT_ADDRESS_TESTNET = '0x887d9102f0003f1724d8fd5d4fe95a11572fcd77';
exports.STARGATE_CONTRACT_ADDRESS_TESTNET = '0x1e02b2953adefec225cf0ec49805b1146a4429c1';
exports.STARGATE_CONTRACT_ADDRESS = '0x03c557be98123fdb6fad325328ac6eb77de7248c';
exports.VALIDATOR_REGISTRATION_STAKER_CONTRACT_ADDRESS_TESTNET = '0x00000000000000000000000000005374616B6572';
exports.VALIDATOR_REGISTRATION_STAKER_CONTRACT_ADDRESS_MAINNET = '0x00000000000000000000000000005374616B6572';
exports.AVG_GAS_UNITS = '21000';
exports.EXPIRATION = 400;
exports.GAS_PRICE_COEF = '128';
exports.GAS_UNIT_PRICE = '10000000000000'; // vechain has fixed gas unit price of 10^13 wei
exports.COEF_DIVISOR = '255';
exports.feeEstimateData = {
    gas: exports.AVG_GAS_UNITS,
    gasUnitPrice: exports.GAS_UNIT_PRICE,
    gasPriceCoef: exports.GAS_PRICE_COEF,
    coefDivisor: exports.COEF_DIVISOR,
};
