// @bitgo/sdk-coin-vet/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transferBuilder_1 = require("./transactionBuilder/transferBuilder");
const addressInitializationBuilder_1 = require("./transactionBuilder/addressInitializationBuilder");
const flushTokenTransactionBuilder_1 = require("./transactionBuilder/flushTokenTransactionBuilder");
const exitDelegationBuilder_1 = require("./transactionBuilder/exitDelegationBuilder");
const burnNftBuilder_1 = require("./transactionBuilder/burnNftBuilder");
const claimRewardsBuilder_1 = require("./transactionBuilder/claimRewardsBuilder");
const transaction_1 = require("./transaction/transaction");
const utils_1 = __importDefault(require("./utils"));
const addressInitializationTransaction_1 = require("./transaction/addressInitializationTransaction");
const flushTokenTransaction_1 = require("./transaction/flushTokenTransaction");
const exitDelegation_1 = require("./transaction/exitDelegation");
const burnNftTransaction_1 = require("./transaction/burnNftTransaction");
const claimRewards_1 = require("./transaction/claimRewards");
const tokenTransactionBuilder_1 = require("./transactionBuilder/tokenTransactionBuilder");
const tokenTransaction_1 = require("./transaction/tokenTransaction");
const stakingBuilder_1 = require("./transactionBuilder/stakingBuilder");
const stakingTransaction_1 = require("./transaction/stakingTransaction");
const nftTransactionBuilder_1 = require("./transactionBuilder/nftTransactionBuilder");
const nftTransaction_1 = require("./transaction/nftTransaction");
const stakeClauseTransaction_1 = require("./transaction/stakeClauseTransaction");
const stakeClauseTxnBuilder_1 = require("./transactionBuilder/stakeClauseTxnBuilder");
const delegateTxnBuilder_1 = require("./transactionBuilder/delegateTxnBuilder");
const delegateClauseTransaction_1 = require("./transaction/delegateClauseTransaction");
const validatorRegistrationTransaction_1 = require("./transaction/validatorRegistrationTransaction");
const validatorRegistrationBuilder_1 = require("./transactionBuilder/validatorRegistrationBuilder");
const increaseStakeTransaction_1 = require("./transaction/increaseStakeTransaction");
const increaseStakeBuilder_1 = require("./transactionBuilder/increaseStakeBuilder");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    from(signedRawTx) {
        try {
            const signedTx = transaction_1.Transaction.deserializeTransaction(signedRawTx);
            const type = this.getTransactionTypeFromSignedTxn(signedTx);
            switch (type) {
                case sdk_core_1.TransactionType.Send:
                    const transferTx = new transaction_1.Transaction(this._coinConfig);
                    transferTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getTransferBuilder(transferTx);
                case sdk_core_1.TransactionType.AddressInitialization:
                    const addressInitializationTx = new addressInitializationTransaction_1.AddressInitializationTransaction(this._coinConfig);
                    addressInitializationTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getAddressInitializationBuilder(addressInitializationTx);
                case sdk_core_1.TransactionType.FlushTokens:
                    const flushTokenTx = new flushTokenTransaction_1.FlushTokenTransaction(this._coinConfig);
                    flushTokenTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getFlushTokenTransactionBuilder(flushTokenTx);
                case sdk_core_1.TransactionType.SendToken:
                    const tokenTransferTx = new tokenTransaction_1.TokenTransaction(this._coinConfig);
                    tokenTransferTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getTokenTransactionBuilder(tokenTransferTx);
                case sdk_core_1.TransactionType.SendNFT:
                    const nftTransferTx = new nftTransaction_1.NFTTransaction(this._coinConfig);
                    nftTransferTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getNFTTransactionBuilder(nftTransferTx);
                case sdk_core_1.TransactionType.ContractCall:
                    const stakingTx = new stakingTransaction_1.StakingTransaction(this._coinConfig);
                    stakingTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getStakingBuilder(stakingTx);
                case sdk_core_1.TransactionType.StakingActivate:
                    const stakeClauseTx = new stakeClauseTransaction_1.StakeClauseTransaction(this._coinConfig);
                    stakeClauseTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getStakingActivateBuilder(stakeClauseTx);
                case sdk_core_1.TransactionType.StakingDelegate:
                    const delegateClauseTx = new delegateClauseTransaction_1.DelegateClauseTransaction(this._coinConfig);
                    delegateClauseTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getStakingDelegateBuilder(delegateClauseTx);
                case sdk_core_1.TransactionType.StakingUnlock:
                    const exitDelegationTx = new exitDelegation_1.ExitDelegationTransaction(this._coinConfig);
                    exitDelegationTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getExitDelegationBuilder(exitDelegationTx);
                case sdk_core_1.TransactionType.StakingWithdraw:
                    const burnNftTx = new burnNftTransaction_1.BurnNftTransaction(this._coinConfig);
                    burnNftTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getBurnNftBuilder(burnNftTx);
                case sdk_core_1.TransactionType.StakingClaim:
                    const claimRewardsTx = new claimRewards_1.ClaimRewardsTransaction(this._coinConfig);
                    claimRewardsTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getClaimRewardsBuilder(claimRewardsTx);
                case sdk_core_1.TransactionType.StakingLock:
                    const validatorRegistrationTx = new validatorRegistrationTransaction_1.ValidatorRegistrationTransaction(this._coinConfig);
                    validatorRegistrationTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getValidatorRegistrationBuilder(validatorRegistrationTx);
                case sdk_core_1.TransactionType.StakingAdd:
                    const increaseStakeTx = new increaseStakeTransaction_1.IncreaseStakeTransaction(this._coinConfig);
                    increaseStakeTx.fromDeserializedSignedTransaction(signedTx);
                    return this.getIncreaseStakeBuilder(increaseStakeTx);
                default:
                    throw new sdk_core_1.InvalidTransactionError('Invalid transaction type');
            }
        }
        catch (e) {
            throw e;
        }
    }
    /** @inheritdoc */
    getTransferBuilder(tx) {
        return this.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    getAddressInitializationBuilder(tx) {
        return this.initializeBuilder(tx, new addressInitializationBuilder_1.AddressInitializationBuilder(this._coinConfig));
    }
    getFlushTokenTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new flushTokenTransactionBuilder_1.FlushTokenTransactionBuilder(this._coinConfig));
    }
    getTokenTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new tokenTransactionBuilder_1.TokenTransactionBuilder(this._coinConfig));
    }
    getStakingBuilder(tx) {
        return this.initializeBuilder(tx, new stakingBuilder_1.StakingBuilder(this._coinConfig));
    }
    getStakingDelegateBuilder(tx) {
        return this.initializeBuilder(tx, new delegateTxnBuilder_1.DelegateTxnBuilder(this._coinConfig));
    }
    getValidatorRegistrationBuilder(tx) {
        return this.initializeBuilder(tx, new validatorRegistrationBuilder_1.ValidatorRegistrationBuilder(this._coinConfig));
    }
    getIncreaseStakeBuilder(tx) {
        return this.initializeBuilder(tx, new increaseStakeBuilder_1.IncreaseStakeBuilder(this._coinConfig));
    }
    getStakingActivateBuilder(tx) {
        return this.initializeBuilder(tx, new stakeClauseTxnBuilder_1.StakeClauseTxnBuilder(this._coinConfig));
    }
    /**
     * Gets an nft transaction builder.
     *
     * @param {NFTTransaction} tx - The nft transaction to use
     * @returns {NFTTransactionBuilder} The nft transaction builder
     */
    getNFTTransactionBuilder(tx) {
        return this.initializeBuilder(tx, new nftTransactionBuilder_1.NFTTransactionBuilder(this._coinConfig));
    }
    /**
     * Gets an exit delegation transaction builder.
     *
     * @param {ExitDelegationTransaction} tx - The exit delegation transaction to use
     * @returns {ExitDelegationBuilder} The exit delegation transaction builder
     */
    getExitDelegationBuilder(tx) {
        return this.initializeBuilder(tx, new exitDelegationBuilder_1.ExitDelegationBuilder(this._coinConfig));
    }
    /**
     * Gets a burn NFT transaction builder.
     *
     * @param {BurnNftTransaction} tx - The burn NFT transaction to use
     * @returns {BurnNftBuilder} The burn NFT transaction builder
     */
    getBurnNftBuilder(tx) {
        return this.initializeBuilder(tx, new burnNftBuilder_1.BurnNftBuilder(this._coinConfig));
    }
    /**
     * Gets a claim rewards transaction builder.
     *
     * @param {ClaimRewardsTransaction} tx - The claim rewards transaction to use
     * @returns {ClaimRewardsBuilder} The claim rewards transaction builder
     */
    getClaimRewardsBuilder(tx) {
        return this.initializeBuilder(tx, new claimRewardsBuilder_1.ClaimRewardsBuilder(this._coinConfig));
    }
    /** @inheritdoc */
    getWalletInitializationBuilder() {
        throw new Error('Method not implemented.');
    }
    getTransactionTypeFromSignedTxn(signedTxn) {
        return utils_1.default.getTransactionTypeFromClause(signedTxn.body.clauses);
    }
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
