// @bitgo/sdk-coin-vet/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const sdk_core_1 = require("@vechain/sdk-core");
const ethereumjs_abi_1 = __importDefault(require("ethereumjs-abi"));
const ethereumjs_util_1 = require("ethereumjs-util");
const sdk_core_2 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const constants_1 = require("./constants");
const keyPair_1 = require("./keyPair");
class Utils {
    isValidAddress(address) {
        return this.isValidHex(address, constants_1.VET_ADDRESS_LENGTH);
    }
    isValidBlockId(hash) {
        return this.isValidHex(hash, constants_1.VET_BLOCK_ID_LENGTH);
    }
    isValidPrivateKey(key) {
        try {
            new keyPair_1.KeyPair({ prv: key });
            return true;
        }
        catch (e) {
            return false;
        }
    }
    isValidPublicKey(key) {
        try {
            new keyPair_1.KeyPair({ pub: key });
            return true;
        }
        catch (e) {
            return false;
        }
    }
    isValidSignature(signature) {
        throw new Error('Method not implemented');
    }
    isValidTransactionId(txId) {
        return this.isValidHex(txId, constants_1.VET_TRANSACTION_ID_LENGTH);
    }
    isValidHex(value, length) {
        const regex = new RegExp(`^(0x|0X)[a-fA-F0-9]{${length}}$`);
        return regex.test(value);
    }
    deserializeTransaction(serializedTransaction) {
        const txBytes = sdk_core_1.HexUInt.of(serializedTransaction).bytes;
        try {
            return sdk_core_1.Transaction.decode(txBytes, false);
        }
        catch (err) {
            if (err.message?.includes('Expected 9 items, but got 10')) {
                // Likely signed, so retry with isSigned = true
                return sdk_core_1.Transaction.decode(txBytes, true);
            }
            throw err;
        }
    }
    getTransactionTypeFromClause(clauses) {
        if (clauses[0].data === '0x') {
            return sdk_core_2.TransactionType.Send;
        }
        else if (clauses[0].data.startsWith(abstract_eth_1.v4CreateForwarderMethodId)) {
            return sdk_core_2.TransactionType.AddressInitialization;
        }
        else if (clauses[0].data.startsWith(abstract_eth_1.flushForwarderTokensMethodIdV4)) {
            return sdk_core_2.TransactionType.FlushTokens;
        }
        else if (clauses[0].data.startsWith(constants_1.TRANSFER_TOKEN_METHOD_ID)) {
            return sdk_core_2.TransactionType.SendToken;
        }
        else if (clauses[0].data.startsWith(constants_1.STAKING_METHOD_ID)) {
            return sdk_core_2.TransactionType.ContractCall;
        }
        else if (clauses[0].data.startsWith(constants_1.STAKE_CLAUSE_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingActivate;
        }
        else if (clauses[0].data.startsWith(constants_1.DELEGATE_CLAUSE_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingDelegate;
        }
        else if (clauses[0].data.startsWith(constants_1.EXIT_DELEGATION_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingUnlock;
        }
        else if (clauses[0].data.startsWith(constants_1.ADD_VALIDATION_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingLock;
        }
        else if (clauses[0].data.startsWith(constants_1.INCREASE_STAKE_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingAdd;
        }
        else if (clauses[0].data.startsWith(constants_1.BURN_NFT_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingWithdraw;
        }
        else if (clauses[0].data.startsWith(constants_1.CLAIM_BASE_REWARDS_METHOD_ID) ||
            clauses[0].data.startsWith(constants_1.CLAIM_STAKING_REWARDS_METHOD_ID)) {
            return sdk_core_2.TransactionType.StakingClaim;
        }
        else if (clauses[0].data.startsWith(constants_1.TRANSFER_NFT_METHOD_ID)) {
            return sdk_core_2.TransactionType.SendNFT;
        }
        else {
            return sdk_core_2.TransactionType.SendToken;
        }
    }
    getTransferTokenData(toAddress, amountWei) {
        const methodName = 'transfer';
        const types = ['address', 'uint256'];
        const params = [toAddress, new ethereumjs_util_1.BN(amountWei)];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    getTransferNFTData(from, to, tokenId) {
        const methodName = 'transferFrom';
        const types = ['address', 'address', 'uint256'];
        const params = [from, to, new ethereumjs_util_1.BN(tokenId)];
        const method = ethereumjs_abi_1.default.methodID(methodName, types);
        const args = ethereumjs_abi_1.default.rawEncode(types, params);
        return (0, ethereumjs_util_1.addHexPrefix)(Buffer.concat([method, args]).toString('hex'));
    }
    /**
     * Decodes staking transaction data to extract levelId and autorenew
     *
     * @param {string} data - The encoded transaction data
     * @returns {object} - Object containing levelId and autorenew
     */
    decodeStakingData(data) {
        try {
            const parameters = data.slice(10);
            // Decode using ethereumjs-abi directly
            const decoded = ethereumjs_abi_1.default.rawDecode(['uint8', 'bool'], Buffer.from(parameters, 'hex'));
            return {
                levelId: Number(decoded[0]),
                autorenew: Boolean(decoded[1]),
            };
        }
        catch (error) {
            throw new Error(`Failed to decode staking data: ${error.message}`);
        }
    }
    /**
     * Decodes staking transaction data to extract levelId and autorenew
     *
     * @param {string} data - The encoded transaction data
     * @returns {object} - Object containing levelId and autorenew
     */
    decodeStakeClauseData(data) {
        try {
            const parameters = data.slice(10);
            // Decode using ethereumjs-abi directly
            const decoded = ethereumjs_abi_1.default.rawDecode(['uint8'], Buffer.from(parameters, 'hex'));
            return {
                levelId: Number(decoded[0]),
            };
        }
        catch (error) {
            throw new Error(`Failed to decode staking data: ${error.message}`);
        }
    }
    /**
     * Decodes delegate transaction data to extract tokenId and validatorAddress
     *
     * @param {string} data - The encoded transaction data
     * @returns {object} - Object containing levelId and validator address
     */
    decodeDelegateClauseData(data) {
        try {
            const parameters = data.slice(10);
            // Decode using ethereumjs-abi directly
            const decoded = ethereumjs_abi_1.default.rawDecode(['uint256', 'address'], Buffer.from(parameters, 'hex'));
            return {
                tokenId: String(decoded[0]),
                validator: (0, ethereumjs_util_1.addHexPrefix)(decoded[1].toString()).toLowerCase(),
            };
        }
        catch (error) {
            throw new Error(`Failed to decode delegation data: ${error.message}`);
        }
    }
    decodeTransferTokenData(data) {
        const [address, amount] = (0, abstract_eth_1.getRawDecoded)(['address', 'uint256'], (0, abstract_eth_1.getBufferedByteCode)(constants_1.TRANSFER_TOKEN_METHOD_ID, data));
        const recipientAddress = (0, ethereumjs_util_1.addHexPrefix)(address.toString()).toLowerCase();
        return {
            address: recipientAddress,
            amount: amount.toString(),
        };
    }
    decodeTransferNFTData(data) {
        const [from, to, tokenIdBN] = (0, abstract_eth_1.getRawDecoded)(['address', 'address', 'uint256'], (0, abstract_eth_1.getBufferedByteCode)(constants_1.TRANSFER_NFT_METHOD_ID, data));
        const recipientAddress = (0, ethereumjs_util_1.addHexPrefix)(to.toString()).toLowerCase();
        const recipient = {
            address: recipientAddress,
            amount: '1',
        };
        const sender = (0, ethereumjs_util_1.addHexPrefix)(from.toString()).toLowerCase();
        const tokenId = tokenIdBN.toString();
        return {
            recipients: [recipient],
            sender,
            tokenId,
        };
    }
    /**
     * Decodes claim rewards transaction data to extract tokenId
     *
     * @param {string} data - The encoded claim rewards method call data
     * @returns {string} - The tokenId as a string
     */
    decodeClaimRewardsData(data) {
        try {
            // Remove method ID (first 10 characters: '0x' + 4-byte method ID)
            const methodData = data.slice(10);
            // Extract tokenId from first 32-byte slot
            // The tokenId is a uint256, so we take the full 64 hex characters
            const tokenIdHex = methodData.slice(0, 64);
            // Convert hex to decimal string
            return BigInt('0x' + tokenIdHex).toString();
        }
        catch (error) {
            throw new Error(`Failed to decode claim rewards data: ${error.message}`);
        }
    }
    /**
     * Decodes add validation transaction data to extract validator address and staking period
     *
     * @param {string} data - The encoded transaction data
     * @returns {object} - Object containing validator address and staking period
     */
    decodeAddValidationData(data) {
        try {
            const parameters = data.slice(10);
            // Decode using ethereumjs-abi directly
            const decoded = ethereumjs_abi_1.default.rawDecode(['address', 'uint32'], Buffer.from(parameters, 'hex'));
            return {
                validator: (0, ethereumjs_util_1.addHexPrefix)(decoded[0].toString()).toLowerCase(),
                period: Number(decoded[1]),
            };
        }
        catch (error) {
            throw new Error(`Failed to decode add validation data: ${error.message}`);
        }
    }
    /**
     * Decodes increase stake transaction data to extract validator address
     *
     * @param {string} data - The encoded transaction data
     * @returns {object} - Object containing validator address
     */
    decodeIncreaseStakeData(data) {
        try {
            const parameters = data.slice(10);
            // Decode using ethereumjs-abi directly
            const decoded = ethereumjs_abi_1.default.rawDecode(['address'], Buffer.from(parameters, 'hex'));
            return {
                validator: (0, ethereumjs_util_1.addHexPrefix)(decoded[0].toString()).toLowerCase(),
            };
        }
        catch (error) {
            throw new Error(`Failed to decode increase stake data: ${error.message}`);
        }
    }
    /**
     * Decodes exit delegation transaction data to extract tokenId
     *
     * @param {string} data - The encoded exit delegation method call data
     * @returns {string} - The tokenId as a string
     */
    decodeExitDelegationData(data) {
        try {
            if (!data.startsWith(constants_1.EXIT_DELEGATION_METHOD_ID)) {
                throw new Error('Invalid exit delegation method data');
            }
            const tokenIdHex = data.slice(constants_1.EXIT_DELEGATION_METHOD_ID.length);
            // Convert hex to decimal (matching original parseInt logic)
            return parseInt(tokenIdHex, 16).toString();
        }
        catch (error) {
            throw new Error(`Failed to decode exit delegation data: ${error.message}`);
        }
    }
    /**
     * Decodes burn NFT transaction data to extract tokenId
     *
     * @param {string} data - The encoded burn NFT method call data
     * @returns {string} - The tokenId as a string
     */
    decodeBurnNftData(data) {
        try {
            if (!data.startsWith(constants_1.BURN_NFT_METHOD_ID)) {
                throw new Error('Invalid burn NFT method data');
            }
            const tokenIdHex = data.slice(constants_1.BURN_NFT_METHOD_ID.length);
            // Convert hex to decimal (matching original parseInt logic)
            return parseInt(tokenIdHex, 16).toString();
        }
        catch (error) {
            throw new Error(`Failed to decode burn NFT data: ${error.message}`);
        }
    }
    /**
     * Get the network-appropriate staking contract address
     * @param {CoinConfig} coinConfig - The coin configuration object
     * @returns {string} The staking contract address for the network
     */
    getDefaultStakingAddress(coinConfig) {
        const isTestnet = coinConfig.network.type === 'testnet';
        return isTestnet ? constants_1.STARGATE_CONTRACT_ADDRESS_TESTNET : constants_1.STARGATE_CONTRACT_ADDRESS;
    }
    /**
     * Get the network-appropriate contract address for validator registration
     * @param {CoinConfig} coinConfig - The coin configuration object
     * @returns {string} The contract address for the network
     */
    getContractAddressForValidatorRegistration(coinConfig) {
        const isTestnet = coinConfig.network.type === 'testnet';
        return isTestnet
            ? constants_1.VALIDATOR_REGISTRATION_STAKER_CONTRACT_ADDRESS_TESTNET
            : constants_1.VALIDATOR_REGISTRATION_STAKER_CONTRACT_ADDRESS_MAINNET;
    }
    /**
     * Check if an address is a valid NFT contract address for any network
     * @param {string} address - The address to check
     * @returns {boolean} True if the address is an NFT contract address
     */
    isNftContractAddress(address) {
        const lowerAddress = address.toLowerCase();
        return (lowerAddress === constants_1.STARGATE_NFT_ADDRESS.toLowerCase() || lowerAddress === constants_1.STARGATE_NFT_ADDRESS_TESTNET.toLowerCase());
    }
    /**
     * Validate that a contract address matches the expected stargate address for the network
     * @param {string} address - The contract address to validate
     * @param {CoinConfig} coinConfig - The coin configuration object
     * @throws {Error} If the address doesn't match the expected contract address
     */
    validateStakingContractAddress(address, coinConfig) {
        const expectedAddress = this.getDefaultStakingAddress(coinConfig);
        if (address.toLowerCase() !== expectedAddress.toLowerCase()) {
            throw new Error(`Invalid staking contract address. Expected ${expectedAddress} for ${coinConfig.network.type}, got ${address}`);
        }
    }
    /**
     * Validate that the contract address matches the expected stargate address for the network
     * @param {string} address - The contract address to validate
     * @param {CoinConfig} coinConfig - The coin configuration object
     * @throws {Error} If the address doesn't match the expected contract address
     */
    validateContractAddressForValidatorRegistration(address, coinConfig) {
        const expectedAddress = this.getContractAddressForValidatorRegistration(coinConfig);
        if (address.toLowerCase() !== expectedAddress.toLowerCase()) {
            throw new Error(`Invalid contract address for validator registration. Expected ${expectedAddress} for ${coinConfig.network.type}, got ${address}`);
        }
    }
}
exports.Utils = Utils;
const utils = new Utils();
exports.default = utils;
