// @bitgo/abstract-utxo/dist/cjs/src/descriptor/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWallet = exports.getPolicyForEnv = exports.getDescriptorMapFromWallet = exports.isDescriptorWallet = exports.toNamedDescriptorNative = exports.hasValidSignature = exports.createNamedDescriptorWithSignature = exports.NamedDescriptor = exports.assertDescriptorWalletAddress = exports.Descriptor = exports.Miniscript = void 0;
var wasm_utxo_1 = require("@bitgo/wasm-utxo");
Object.defineProperty(exports, "Miniscript", { enumerable: true, get: function () { return wasm_utxo_1.Miniscript; } });
Object.defineProperty(exports, "Descriptor", { enumerable: true, get: function () { return wasm_utxo_1.Descriptor; } });
var assertDescriptorWalletAddress_1 = require("./assertDescriptorWalletAddress");
Object.defineProperty(exports, "assertDescriptorWalletAddress", { enumerable: true, get: function () { return assertDescriptorWalletAddress_1.assertDescriptorWalletAddress; } });
var NamedDescriptor_1 = require("./NamedDescriptor");
Object.defineProperty(exports, "NamedDescriptor", { enumerable: true, get: function () { return NamedDescriptor_1.NamedDescriptor; } });
Object.defineProperty(exports, "createNamedDescriptorWithSignature", { enumerable: true, get: function () { return NamedDescriptor_1.createNamedDescriptorWithSignature; } });
Object.defineProperty(exports, "hasValidSignature", { enumerable: true, get: function () { return NamedDescriptor_1.hasValidSignature; } });
Object.defineProperty(exports, "toNamedDescriptorNative", { enumerable: true, get: function () { return NamedDescriptor_1.toNamedDescriptorNative; } });
var descriptorWallet_1 = require("./descriptorWallet");
Object.defineProperty(exports, "isDescriptorWallet", { enumerable: true, get: function () { return descriptorWallet_1.isDescriptorWallet; } });
Object.defineProperty(exports, "getDescriptorMapFromWallet", { enumerable: true, get: function () { return descriptorWallet_1.getDescriptorMapFromWallet; } });
var validatePolicy_1 = require("./validatePolicy");
Object.defineProperty(exports, "getPolicyForEnv", { enumerable: true, get: function () { return validatePolicy_1.getPolicyForEnv; } });
exports.createWallet = __importStar(require("./createWallet"));
