// @bitgo/abstract-utxo/dist/cjs/src/descriptor/NamedDescriptor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamedDescriptor = void 0;
exports.createNamedDescriptorWithSignature = createNamedDescriptorWithSignature;
exports.toNamedDescriptorNative = toNamedDescriptorNative;
exports.hasValidSignature = hasValidSignature;
exports.assertHasValidSignature = assertHasValidSignature;
const t = __importStar(require("io-ts"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
exports.NamedDescriptor = t.intersection([
    t.type({
        name: t.string,
        value: t.string,
    }),
    t.partial({
        signatures: t.union([t.array(t.string), t.undefined]),
    }),
], 'NamedDescriptor');
function createNamedDescriptorWithSignature(name, descriptor, signingKey) {
    if (typeof descriptor === 'string') {
        descriptor = wasm_utxo_1.Descriptor.fromString(descriptor, 'derivable');
    }
    const value = descriptor.toString();
    const signature = Buffer.from(wasm_utxo_1.message.signMessage(value, signingKey.privateKey)).toString('hex');
    return { name, value, signatures: [signature] };
}
function toNamedDescriptorNative(e, pkType) {
    return { ...e, value: wasm_utxo_1.Descriptor.fromString(e.value, pkType) };
}
function hasValidSignature(descriptor, key, signatures) {
    if (typeof descriptor === 'string') {
        descriptor = wasm_utxo_1.Descriptor.fromString(descriptor, 'derivable');
    }
    const descriptorString = descriptor.toString();
    return signatures.some((signature) => {
        return wasm_utxo_1.message.verifyMessage(descriptorString, key.publicKey, Buffer.from(signature, 'hex'));
    });
}
function assertHasValidSignature(namedDescriptor, key) {
    if (!hasValidSignature(namedDescriptor.value, key, namedDescriptor.signatures ?? [])) {
        throw new Error(`Descriptor ${namedDescriptor.name} does not have a valid signature (key=${key.toBase58()})`);
    }
}
