// @bitgo/abstract-utxo/dist/cjs/src/descriptor/validatePolicy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescriptorPolicyValidationError = exports.policyAllowAll = void 0;
exports.getValidatorDescriptorTemplate = getValidatorDescriptorTemplate;
exports.getValidatorEvery = getValidatorEvery;
exports.getValidatorSome = getValidatorSome;
exports.getValidatorOneOfTemplates = getValidatorOneOfTemplates;
exports.getValidatorSignedByUserKey = getValidatorSignedByUserKey;
exports.assertDescriptorPolicy = assertDescriptorPolicy;
exports.toDescriptorMapValidate = toDescriptorMapValidate;
exports.getPolicyForEnv = getPolicyForEnv;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const builder_1 = require("./builder");
const NamedDescriptor_1 = require("./NamedDescriptor");
exports.policyAllowAll = {
    name: 'allowAll',
    validate: () => true,
};
function getValidatorDescriptorTemplate(name) {
    return {
        name: 'descriptorTemplate(' + name + ')',
        validate(arr, walletKeys) {
            return arr.every((d) => {
                const parsed = (0, builder_1.parseDescriptor)(d.value);
                return (parsed.name === name &&
                    parsed.keys.length === walletKeys.length &&
                    parsed.keys.every((k, i) => k.toBase58() === walletKeys[i].neutered().toBase58()));
            });
        },
    };
}
function getValidatorEvery(validators) {
    return {
        name: 'every(' + validators.map((v) => v.name).join(',') + ')',
        validate(arr, walletKeys) {
            return validators.every((v) => v.validate(arr, walletKeys));
        },
    };
}
function getValidatorSome(validators) {
    return {
        name: 'some(' + validators.map((v) => v.name).join(',') + ')',
        validate(arr, walletKeys) {
            return validators.some((v) => v.validate(arr, walletKeys));
        },
    };
}
function getValidatorOneOfTemplates(names) {
    return getValidatorSome(names.map(getValidatorDescriptorTemplate));
}
function getValidatorSignedByUserKey() {
    return {
        name: 'signedByUser',
        validate(arr, walletKeys) {
            // the first key is the user key, by convention
            return arr.every((d) => (0, NamedDescriptor_1.hasValidSignature)(d.value, walletKeys[0], d.signatures ?? []));
        },
    };
}
class DescriptorPolicyValidationError extends Error {
    constructor(ds, policy) {
        super(`Descriptors ${ds.map((d) => d.value.toString())} does not match policy ${policy.name}`);
    }
}
exports.DescriptorPolicyValidationError = DescriptorPolicyValidationError;
function assertDescriptorPolicy(descriptors, policy, walletKeys) {
    if (!policy.validate(descriptors, walletKeys)) {
        throw new DescriptorPolicyValidationError(descriptors, policy);
    }
}
function toDescriptorMapValidate(descriptors, walletKeys, policy) {
    const namedDescriptorsNative = descriptors.map((v) => (0, NamedDescriptor_1.toNamedDescriptorNative)(v, 'derivable'));
    assertDescriptorPolicy(namedDescriptorsNative, policy, walletKeys);
    return wasm_utxo_1.descriptorWallet.toDescriptorMap(namedDescriptorsNative);
}
function getPolicyForEnv(env) {
    switch (env) {
        case 'adminProd':
        case 'prod':
            return getValidatorSome([
                // allow 2-of-3-ish descriptor groups where the keys match the wallet keys
                getValidatorDescriptorTemplate('Wsh2Of3'),
                // allow descriptor groups where all keys match the wallet keys plus OP_DROP (coredao staking)
                getValidatorDescriptorTemplate('Wsh2Of3CltvDrop'),
                // allow all descriptors signed by the user key
                getValidatorSignedByUserKey(),
            ]);
        default:
            return exports.policyAllowAll;
    }
}
