// @bitgo/public-types/dist/src/schema/transactionRequest/intents/accountBaseConsolidateIntent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountBaseConsolidateIntent = void 0;
const t = __importStar(require("io-ts"));
const accountBaseBuildOptions_1 = require("./accountBaseBuildOptions");
const intentType_1 = require("./intentType");
const utils_1 = require("../../../utils");
const tokenRecipientEntry_1 = require("./tokenRecipientEntry");
exports.AccountBaseConsolidateIntent = t.intersection([
    accountBaseBuildOptions_1.AccountBaseBuildOptions,
    t.type({
        intentType: intentType_1.intentTypes.consolidate,
        receiveAddress: t.string,
    }),
    t.partial({
        senderAddressIndex: utils_1.optionalNumber,
        consolidateId: utils_1.optionalString,
        recipients: (0, utils_1.Optional)(t.array(tokenRecipientEntry_1.TokenRecipientEntry)),
        keepAlive: utils_1.optionalBoolean,
    }),
]);
