// @bitgo/public-types/dist/src/schema/transactionRequest/intents/aptCustomTxIntent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AptCustomTxIntent = void 0;
const t = __importStar(require("io-ts"));
const aptBuildOptions_1 = require("./aptBuildOptions");
const baseIntent_1 = require("./baseIntent");
const intentType_1 = require("./intentType");
const EntryFunctionABI = t.type({
    typeParameters: t.array(t.any),
    parameters: t.array(t.any),
});
exports.AptCustomTxIntent = t.intersection([
    baseIntent_1.BaseIntent,
    aptBuildOptions_1.AptBuildOptions,
    t.type({
        intentType: intentType_1.intentTypes.customTx,
        aptosCustomTransactionParams: t.intersection([
            t.type({
                moduleName: t.string,
                functionName: t.string,
            }),
            t.partial({
                typeArguments: t.array(t.string),
                functionArguments: t.array(t.any),
                abi: EntryFunctionABI,
            }),
        ]),
    }),
]);
