// @bitgo/public-types/dist/src/schema/transactionRequest/intents/baseIntent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseIntentWithAmount = exports.BaseIntentWithTokenName = exports.BaseIntentWithTokenData = exports.BaseIntentWithTokenRecipients = exports.BaseIntentWithRecipients = exports.BaseIntentWithoutNonce = exports.BaseIntent = void 0;
const t = __importStar(require("io-ts"));
const intentType_1 = require("./intentType");
const utils_1 = require("../../../utils");
const recipientEntry_1 = require("./recipientEntry");
const tokenRecipientEntry_1 = require("./tokenRecipientEntry");
const amount_1 = require("./amount");
exports.BaseIntent = t.intersection([
    t.type({
        intentType: intentType_1.IntentType,
    }),
    t.partial({
        sequenceId: utils_1.optionalString,
        comment: utils_1.optionalString,
        nonce: (0, utils_1.Optional)(t.union([t.string, t.number])),
    }),
]);
exports.BaseIntentWithoutNonce = t.intersection([
    t.type({
        intentType: intentType_1.IntentType,
    }),
    t.partial({
        sequenceId: utils_1.optionalString,
        comment: utils_1.optionalString,
    }),
]);
exports.BaseIntentWithRecipients = t.type({
    recipients: t.array(recipientEntry_1.RecipientEntry),
});
exports.BaseIntentWithTokenRecipients = t.type({
    recipients: t.array(tokenRecipientEntry_1.TokenRecipientEntry),
});
exports.BaseIntentWithTokenData = t.type({
    recipients: t.array(t.partial({
        tokenData: t.partial({
            tokenName: t.string,
        }),
    })),
});
exports.BaseIntentWithTokenName = t.type({
    tokenName: t.string,
});
exports.BaseIntentWithAmount = t.type({
    amount: amount_1.Amount,
});
