// @bitgo/public-types/dist/src/schema/transactionRequest/intents/ethBuildOptions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllEthBuildOptions = exports.EthFillNonceBuildOptions = exports.EthBuildOptionsWithoutFeeOptions = exports.EthBuildOptions = void 0;
const t = __importStar(require("io-ts"));
const accountBaseBuildOptions_1 = require("./accountBaseBuildOptions");
const utils_1 = require("../../../utils");
const feeOption_1 = require("./feeOption");
const eip1559FeeOptions_1 = require("./eip1559FeeOptions");
const hopParams_1 = require("./hopParams");
exports.EthBuildOptions = t.intersection([
    accountBaseBuildOptions_1.AccountBaseBuildOptions,
    t.partial({
        feeOptions: (0, utils_1.Optional)(t.union([feeOption_1.FeeOption, eip1559FeeOptions_1.EIP1559FeeOptions])),
        hopParams: (0, utils_1.Optional)(hopParams_1.HopParams),
        receiveAddress: utils_1.optionalString,
        senderAddressIndex: utils_1.optionalNumber,
    }),
]);
exports.EthBuildOptionsWithoutFeeOptions = t.intersection([
    accountBaseBuildOptions_1.AccountBaseBuildOptions,
    t.partial({
        hopParams: (0, utils_1.Optional)(hopParams_1.HopParams),
        receiveAddress: utils_1.optionalString,
        senderAddressIndex: utils_1.optionalNumber,
    }),
]);
exports.EthFillNonceBuildOptions = t.partial({
    feeOptions: (0, utils_1.Optional)(t.union([feeOption_1.FeeOption, eip1559FeeOptions_1.EIP1559FeeOptions])),
    receiveAddress: utils_1.optionalString,
    senderAddressIndex: utils_1.optionalNumber,
});
exports.AllEthBuildOptions = t.union([
    exports.EthBuildOptions,
    exports.EthBuildOptionsWithoutFeeOptions,
    exports.EthFillNonceBuildOptions,
]);
