// @bitgo/public-types/dist/src/schema/transactionRequest/intents/feeConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeeConfig = void 0;
const t = __importStar(require("io-ts"));
var FeeDimensions;
(function (FeeDimensions) {
    FeeDimensions[FeeDimensions["Bandwidth"] = 0] = "Bandwidth";
    FeeDimensions[FeeDimensions["DBRead"] = 1] = "DBRead";
    FeeDimensions[FeeDimensions["DBWrite"] = 2] = "DBWrite";
    FeeDimensions[FeeDimensions["Compute"] = 3] = "Compute";
})(FeeDimensions || (FeeDimensions = {}));
exports.FeeConfig = t.type({
    weights: t.record(t.string, t.number),
    maxCapacity: t.bigint,
    maxPerSecond: t.bigint,
    targetPerSecond: t.bigint,
    minPrice: t.bigint,
    excessConversionConstant: t.bigint,
});
