// @bitgo/public-types/dist/src/schema/transactionRequest/intents/intentType.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntentType = exports.intentTypes = void 0;
const t = __importStar(require("io-ts"));
exports.intentTypes = {
    payment: t.literal("payment"),
    transferToken: t.literal("transferToken"),
    consolidate: t.literal("consolidate"),
    consolidateToken: t.literal("consolidateToken"),
    fanout: t.literal("fanout"),
    stake: t.literal("stake"),
    unstake: t.literal("unstake"),
    delegate: t.literal("delegate"),
    undelegate: t.literal("undelegate"),
    switchValidator: t.literal("switchValidator"),
    claim: t.literal("claim"),
    stakeClaimRewards: t.literal("stakeClaimRewards"),
    pledge: t.literal("pledge"),
    voteDelegation: t.literal("voteDelegation"),
    vote: t.literal("vote"),
    createAccount: t.literal("createAccount"),
    updateAccount: t.literal("updateAccount"),
    addTrustLine: t.literal("addTrustLine"),
    removeTrustLine: t.literal("removeTrustLine"),
    signMessage: t.literal("signMessage"),
    signTypedStructuredData: t.literal("signTypedStructuredData"),
    enableToken: t.literal("enableToken"),
    authorize: t.literal("authorize"),
    acceleration: t.literal("acceleration"),
    fillNonce: t.literal("fillNonce"),
    walletRecovery: t.literal("walletRecovery"),
    contractCall: t.literal("contractCall"),
    deactivate: t.literal("deactivate"),
    customTx: t.literal("customTx"),
    closeAssociatedTokenAccount: t.literal("closeAssociatedTokenAccount"),
    stakeWithCallData: t.literal("stakeWithCallData"),
    unstakeWithCallData: t.literal("unstakeWithCallData"),
    switchValidatorWithCallData: t.literal("switchValidatorWithCallData"),
    feeAddressTransfer: t.literal("feeAddressTransfer"),
    tokenApproval: t.literal("tokenApproval"),
    transferStake: t.literal("transferStake"),
    validatorRegistration: t.literal("validatorRegistration"),
    spotTransfer: t.literal("spotTransfer"),
    bridgeFunds: t.literal("bridgeFunds"),
    goUnstake: t.literal("goUnstake"),
    createBtcDelegation: t.literal("createBtcDelegation"),
    transferAccept: t.literal("transferAccept"),
    transferReject: t.literal("transferReject"),
    transferAcknowledge: t.literal("transferAcknowledge"),
    transferOfferWithdrawn: t.literal("transferOfferWithdrawn"),
};
exports.IntentType = t.keyof(exports.intentTypes);
