// @bitgo/public-types/dist/src/schema/transactionRequest/intents/solStakeIntent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolJitoStakeIntent = exports.SolStakeIntent = void 0;
const t = __importStar(require("io-ts"));
const solBuildOptions_1 = require("./solBuildOptions");
const stakeIntent_1 = require("./stakeIntent");
const baseIntent_1 = require("./baseIntent");
const utils_1 = require("../../../utils");
const solStakingType_1 = require("./solStakingType");
exports.SolStakeIntent = t.intersection([
    stakeIntent_1.StakeIntent,
    solBuildOptions_1.SolBuildOptions,
    baseIntent_1.BaseIntentWithAmount,
    t.type({
        validatorAddress: t.string,
        isMarinade: (0, utils_1.Optional)(t.boolean),
        stakingType: (0, utils_1.Optional)(solStakingType_1.SolStakingType),
    }),
]);
exports.SolJitoStakeIntent = t.intersection([
    exports.SolStakeIntent,
    t.type({
        stakingType: t.literal("JITO"),
        stakePoolConfig: t.type({
            poolMint: t.string,
            reserveStake: t.string,
            managerFeeAccount: t.string,
        }),
    }),
]);
