// @bitgo/public-types/dist/src/schema/transactionRequest/intents/solUnstakeIntent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolJitoUnstakeIntent = exports.SolUnstakeIntent = void 0;
const t = __importStar(require("io-ts"));
const solBuildOptions_1 = require("./solBuildOptions");
const amount_1 = require("./amount");
const unstakeIntent_1 = require("./unstakeIntent");
const utils_1 = require("../../../utils");
const recipientEntry_1 = require("./recipientEntry");
const solStakingType_1 = require("./solStakingType");
exports.SolUnstakeIntent = t.intersection([
    unstakeIntent_1.UnstakeIntent,
    solBuildOptions_1.SolBuildOptions,
    t.type({
        stakingAddress: t.string,
    }),
    t.partial({
        amount: (0, utils_1.Optional)(amount_1.Amount),
        remainingStakingAmount: (0, utils_1.Optional)(amount_1.Amount),
        isMarinade: (0, utils_1.Optional)(t.boolean),
        stakingType: (0, utils_1.Optional)(solStakingType_1.SolStakingType),
        recipients: (0, utils_1.Optional)(t.array(recipientEntry_1.RecipientEntry)),
    }),
]);
exports.SolJitoUnstakeIntent = t.intersection([
    exports.SolUnstakeIntent,
    t.type({
        stakingType: t.literal("JITO"),
        stakePoolConfig: t.type({
            poolMint: t.string,
            validatorListAccount: t.string,
            managerFeeAccount: t.string,
        }),
        validatorAddress: t.string,
    }),
]);
