// @bitgo/public-types/dist/src/schema/transactionRequest/intents/solVersionedCustomTxIntent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolVersionedCustomTxIntent = void 0;
const t = __importStar(require("io-ts"));
const solBuildOptions_1 = require("./solBuildOptions");
const baseIntent_1 = require("./baseIntent");
const intentType_1 = require("./intentType");
exports.SolVersionedCustomTxIntent = t.intersection([
    baseIntent_1.BaseIntent,
    solBuildOptions_1.SolBuildOptions,
    t.type({
        intentType: intentType_1.intentTypes.customTx,
        solVersionedTransactionData: t.intersection([
            t.type({
                versionedInstructions: t.array(t.type({
                    programIdIndex: t.number,
                    accountKeyIndexes: t.array(t.number),
                    data: t.string,
                })),
                addressLookupTables: t.array(t.type({
                    accountKey: t.string,
                    writableIndexes: t.array(t.number),
                    readonlyIndexes: t.array(t.number),
                })),
                staticAccountKeys: t.array(t.string),
                messageHeader: t.type({
                    numRequiredSignatures: t.number,
                    numReadonlySignedAccounts: t.number,
                    numReadonlyUnsignedAccounts: t.number,
                }),
            }),
            t.partial({
                recentBlockhash: t.string,
            }),
        ]),
    }),
]);
