// @bitgo/sdk-coin-polyx/dist/src/lib/batchUnstakingBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchUnstakingBuilder = void 0;
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const baseBuilder_1 = require("./baseBuilder");
const txnSchema_1 = require("./txnSchema");
const utils_1 = __importDefault(require("./utils"));
class BatchUnstakingBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * Unbond tokens and chill (stop nominating validators)
     *
     * @returns {UnsignedTransaction} an unsigned Polyx transaction
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        const chillCall = txwrapper_polkadot_1.methods.staking.chill({}, baseTxInfo.baseTxInfo, baseTxInfo.options);
        const unbondCall = txwrapper_polkadot_1.methods.staking.unbond({
            value: this._amount,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
        // Create batch all transaction (atomic execution)
        return txwrapper_polkadot_1.methods.utility.batchAll({
            calls: [chillCall.method, unbondCall.method],
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Batch;
    }
    /**
     * The amount to unstake.
     *
     * @param {string} amount
     * @returns {BatchUnstakingBuilder} This unstake builder.
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     * Get the amount to unstake
     */
    getAmount() {
        return this._amount;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn) {
        const methodName = decodedTxn.method?.name;
        if (methodName === 'batchAll') {
            const txMethod = decodedTxn.method.args;
            const calls = txMethod.calls;
            if (calls.length !== 2) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch unstaking transaction: expected 2 calls but got ${calls.length}`);
            }
            // Check that first call is chill
            const firstCallMethod = utils_1.default.decodeMethodName(calls[0], this._registry);
            if (firstCallMethod !== 'chill') {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch unstaking transaction: first call should be chill but got ${firstCallMethod}`);
            }
            // Check that second call is unbond
            const secondCallMethod = utils_1.default.decodeMethodName(calls[1], this._registry);
            if (secondCallMethod !== 'unbond') {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch unstaking transaction: second call should be unbond but got ${secondCallMethod}`);
            }
            // Validate unbond amount - handle both string and number formats
            const unbondArgs = calls[1].args;
            const valueString = typeof unbondArgs.value === 'string' ? unbondArgs.value : unbondArgs.value.toString();
            const validationResult = txnSchema_1.BatchUnstakingTransactionSchema.validate({
                value: valueString,
            });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch unstaking transaction: ${validationResult.error.message}`);
            }
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction type: ${methodName}. Expected batchAll`);
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method && this._method.name === 'batchAll') {
            const txMethod = this._method.args;
            const calls = txMethod.calls;
            if (calls && calls.length === 2) {
                const secondCallMethod = utils_1.default.decodeMethodName(calls[1], this._registry);
                if (secondCallMethod === 'unbond') {
                    const unbondArgs = calls[1].args;
                    const valueString = typeof unbondArgs.value === 'string' ? unbondArgs.value : unbondArgs.value.toString();
                    this.amount(valueString);
                }
            }
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected batchAll`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._amount);
    }
    validateFields(value) {
        const validationResult = txnSchema_1.BatchUnstakingTransactionSchema.validate({
            value,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Batch Unstaking Builder Transaction validation failed: ${validationResult.error.message}`);
        }
    }
    /**
     * Validates fields for testing
     */
    testValidateFields() {
        this.validateFields(this._amount);
    }
}
exports.BatchUnstakingBuilder = BatchUnstakingBuilder;
