// @bitgo/sdk-coin-polyx/dist/src/lib/batchStakingBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchStakingBuilder = void 0;
const baseBuilder_1 = require("./baseBuilder");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const sdk_core_1 = require("@bitgo/sdk-core");
const txnSchema_1 = require("./txnSchema");
const utils_1 = __importDefault(require("./utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class BatchStakingBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        // For nominate operation
        this._validators = [];
        this.material(utils_1.default.getMaterial(_coinConfig.network.type));
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Batch;
    }
    /**
     * Build a batch transaction that combines bond and nominate operations
     * Both operations are required and always atomic (using batchAll)
     */
    buildTransaction() {
        // Ensure both bond and nominate operations are included
        if (!this._amount || this._validators.length === 0) {
            throw new sdk_core_1.InvalidTransactionError('Batch transaction must include both bond and nominate operations');
        }
        const baseTxInfo = this.createBaseTxInfo();
        // Create the individual calls
        const calls = [];
        // Add bond call
        const bondCall = txwrapper_polkadot_1.methods.staking.bond({
            controller: this._controller || this._sender,
            value: this._amount,
            payee: this._payee || 'Staked',
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
        calls.push(bondCall.method);
        // Add nominate call
        const nominateCall = txwrapper_polkadot_1.methods.staking.nominate({
            targets: this._validators,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
        calls.push(nominateCall.method);
        // Always use batchAll (atomic)
        return txwrapper_polkadot_1.methods.utility.batchAll({
            calls,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    }
    /**
     * Set the staking amount for bond
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     * Get the staking amount
     */
    getAmount() {
        return this._amount;
    }
    /**
     * Set the controller account for bond
     */
    controller(controller) {
        this.validateAddress(controller);
        this._controller = controller.address;
        return this;
    }
    /**
     * Get the controller address
     */
    getController() {
        return this._controller;
    }
    /**
     * Set the rewards destination for bond ('Staked', 'Stash','Controller', or { Account: string })
     */
    payee(payee) {
        if (typeof payee === 'object' && payee.Account) {
            this._payee = payee;
        }
        else {
            this._payee = payee;
        }
        return this;
    }
    /**
     * Get the payee
     */
    getPayee() {
        return this._payee;
    }
    /**
     * Set the validators to nominate
     */
    validators(validators) {
        for (const address of validators) {
            this.validateAddress({ address });
        }
        this._validators = validators;
        return this;
    }
    /**
     * Get the validators to nominate
     */
    getValidators() {
        return this._validators;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn) {
        const methodName = decodedTxn.method?.name;
        // batch bond and nominate
        if (methodName === 'batchAll') {
            const txMethod = decodedTxn.method.args;
            const calls = txMethod.calls;
            if (calls.length !== 2) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch staking transaction: expected 2 calls but got ${calls.length}`);
            }
            // Check that first call is bond
            const firstCallMethod = utils_1.default.decodeMethodName(calls[0], this._registry);
            if (firstCallMethod !== 'bond') {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch staking transaction: first call should be bond but got ${firstCallMethod}`);
            }
            // Check that second call is nominate
            const secondCallMethod = utils_1.default.decodeMethodName(calls[1], this._registry);
            if (secondCallMethod !== 'nominate') {
                throw new sdk_core_1.InvalidTransactionError(`Invalid batch staking transaction: second call should be nominate but got ${secondCallMethod}`);
            }
            // Validate bond arguments
            const bondArgs = calls[0].args;
            this.validateBondArgs(bondArgs);
            // Validate nominate arguments
            const nominateArgs = calls[1].args;
            this.validateNominateArgs(nominateArgs);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction type: ${methodName}`);
        }
    }
    /**
     * Validate bond arguments
     */
    validateBondArgs(args) {
        // Handle both string and object formats for controller
        const controllerValue = args.controller;
        const controllerAddress = typeof controllerValue === 'string' ? controllerValue : controllerValue.id;
        if (!utils_1.default.isValidAddress(controllerAddress)) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid bond args: controller address ${controllerAddress} is not a well-formed address`);
        }
        // Handle both string and number formats for value
        const amountValue = args.value;
        const valueString = typeof amountValue === 'string' ? amountValue : amountValue.toString();
        // Handle different payee formats
        const payeeValue = args.payee;
        let normalizedPayee = payeeValue;
        if (typeof payeeValue === 'object' && payeeValue !== null) {
            const decodedPayee = payeeValue;
            if (decodedPayee.staked !== undefined) {
                normalizedPayee = 'Staked';
            }
            else if (decodedPayee.stash !== undefined) {
                normalizedPayee = 'Stash';
            }
            else if (decodedPayee.controller !== undefined) {
                normalizedPayee = 'Controller';
            }
            else if (decodedPayee.account) {
                normalizedPayee = { Account: decodedPayee.account };
            }
        }
        const validationResult = txnSchema_1.BatchTransactionSchema.validateBond({
            value: valueString,
            controller: controllerAddress,
            payee: normalizedPayee,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid bond args: ${validationResult.error.message}`);
        }
    }
    /**
     * Validate nominate arguments
     */
    validateNominateArgs(args) {
        // Handle both string and object formats for targets
        const targetAddresses = args.targets.map((target) => {
            if (typeof target === 'string') {
                return target;
            }
            else if (target && typeof target === 'object' && 'id' in target) {
                return target.id;
            }
            throw new sdk_core_1.InvalidTransactionError(`Invalid target format: ${JSON.stringify(target)}`);
        });
        const validationResult = txnSchema_1.BatchTransactionSchema.validateNominate({
            validators: targetAddresses,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid nominate args: ${validationResult.error.message}`);
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        // Check if the transaction is a batch transaction
        if (this._method?.name !== 'batchAll') {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected batchAll`);
        }
        if (this._method) {
            const txMethod = this._method.args;
            for (const call of txMethod.calls) {
                const callMethod = utils_1.default.decodeMethodName(call, this._registry);
                if (callMethod === 'bond') {
                    const bondArgs = call.args;
                    // Handle both string and number formats for value
                    const amountValue = bondArgs.value;
                    const valueString = typeof amountValue === 'string' ? amountValue : amountValue.toString();
                    this.amount(valueString);
                    // Handle both string and object formats for controller
                    const controllerValue = bondArgs.controller;
                    const controllerAddress = typeof controllerValue === 'string' ? controllerValue : controllerValue.id;
                    this.controller({ address: controllerAddress });
                    // Handle different payee formats
                    const payeeValue = bondArgs.payee;
                    let normalizedPayee = payeeValue;
                    if (typeof payeeValue === 'object' && payeeValue !== null) {
                        const decodedPayee = payeeValue;
                        if (decodedPayee.staked !== undefined) {
                            normalizedPayee = 'Staked';
                        }
                        else if (decodedPayee.stash !== undefined) {
                            normalizedPayee = 'Stash';
                        }
                        else if (decodedPayee.controller !== undefined) {
                            normalizedPayee = 'Controller';
                        }
                        else if (decodedPayee.account) {
                            normalizedPayee = { Account: decodedPayee.account };
                        }
                    }
                    this.payee(normalizedPayee);
                }
                else if (callMethod === 'nominate') {
                    const nominateArgs = call.args;
                    // Handle both string and object formats for targets
                    const targetAddresses = nominateArgs.targets.map((target) => {
                        if (typeof target === 'string') {
                            return target;
                        }
                        else if (target && typeof target === 'object' && 'id' in target) {
                            return target.id;
                        }
                        throw new sdk_core_1.InvalidTransactionError(`Invalid target format: ${JSON.stringify(target)}`);
                    });
                    this.validators(targetAddresses);
                }
            }
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(tx) {
        super.validateTransaction(tx);
        this.validateFields();
    }
    /**
     * Validate the builder fields
     */
    validateFields() {
        // Ensure both bond and nominate operations are included
        if (!this._amount || this._validators.length === 0) {
            throw new sdk_core_1.InvalidTransactionError('Batch transaction must include both bond and nominate operations');
        }
        const validationResult = txnSchema_1.BatchTransactionSchema.validate({
            amount: this._amount,
            controller: this._controller,
            payee: this._payee,
            validators: this._validators,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction: ${validationResult.error.message}`);
        }
    }
    testValidateFields() {
        this.validateFields();
    }
    testValidateBondArgs(args) {
        return this.validateBondArgs(args);
    }
    testValidateNominateArgs(args) {
        return this.validateNominateArgs(args);
    }
}
exports.BatchStakingBuilder = BatchStakingBuilder;
