// @bitgo/sdk-coin-polyx/dist/src/lib/bondExtraBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BondExtraBuilder = void 0;
const baseBuilder_1 = require("./baseBuilder");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const sdk_core_1 = require("@bitgo/sdk-core");
const txnSchema_1 = require("./txnSchema");
const utils_1 = __importDefault(require("./utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class BondExtraBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.material(utils_1.default.getMaterial(_coinConfig.network.type));
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingActivate;
    }
    /**
     * Build the bondExtra transaction
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return txwrapper_polkadot_1.methods.staking.bondExtra({
            maxAdditional: this._amount,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    }
    /**
     * Set additional amount to stake
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     * Get the amount to stake
     */
    getAmount() {
        return this._amount;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn) {
        const methodName = decodedTxn.method?.name;
        if (methodName === 'bondExtra') {
            const txMethod = decodedTxn.method.args;
            const value = txMethod.maxAdditional;
            const validationResult = txnSchema_1.BondExtraTransactionSchema.validate({
                value,
            });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid transaction: ${validationResult.error.message}`);
            }
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction type: ${methodName}`);
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        const methodName = this._method?.name;
        if (methodName === 'bondExtra' && this._method) {
            const txMethod = this._method.args;
            this.amount(txMethod.maxAdditional);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${methodName}. Expected bondExtra`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(tx) {
        super.validateTransaction(tx);
        this.validateFields();
    }
    /**
     * Validate the bondExtra fields
     */
    validateFields() {
        const validationResult = txnSchema_1.BondExtraTransactionSchema.validate({
            value: this._amount,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction: ${validationResult.error.message}`);
        }
    }
    /**
     * Validate amount
     */
    validateAmount(amount) {
        const amountBN = new bignumber_js_1.default(amount);
        if (amountBN.isNaN() || amountBN.isLessThanOrEqualTo(0)) {
            throw new Error(`Bond amount ${amount} must be a positive number`);
        }
    }
}
exports.BondExtraBuilder = BondExtraBuilder;
