// @bitgo/sdk-coin-polyx/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodNames = exports.PortfolioKind = exports.SettlementType = void 0;
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
/**
 * Settlement type for Polyx transactions
 */
var SettlementType;
(function (SettlementType) {
    SettlementType["SettleOnAffirmation"] = "SettleOnAffirmation";
})(SettlementType || (exports.SettlementType = SettlementType = {}));
/**
 * Portfolio kind for Polyx transactions
 */
var PortfolioKind;
(function (PortfolioKind) {
    PortfolioKind["Default"] = "Default";
})(PortfolioKind || (exports.PortfolioKind = PortfolioKind = {}));
/**
 * Method names for Polyx transactions.
 * Extends the base MethodNames from Interface with additional Polyx-specific methods.
 */
exports.MethodNames = {
    // Include all values from the base object
    ...abstract_substrate_1.Interface.MethodNames,
    /**
     * Registers a Decentralized Identifier (DID) along with Customer Due Diligence (CDD) information.
     *
     * @see https://developers.polymesh.network/sdk-docs/enums/Generated/Types/IdentityTx/#cddregisterdidwithcdd
     */
    RegisterDidWithCDD: 'cddRegisterDidWithCdd',
    /**
     * Pre-approves an asset.
     */
    PreApproveAsset: 'preApproveAsset',
    AddAndAffirmWithMediators: 'addAndAffirmWithMediators',
    RejectInstruction: 'rejectInstruction',
};
