// @bitgo/sdk-coin-polyx/dist/src/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidDid = exports.utils = exports.Utils = exports.WithdrawUnbondedBuilder = exports.UnbondBuilder = exports.BatchUnstakingBuilder = exports.BatchBuilder = exports.BondExtraBuilder = exports.PolyxTransaction = exports.RejectInstructionBuilder = exports.TokenTransferBuilder = exports.PreApproveAssetBuilder = exports.RegisterDidWithCDDBuilder = exports.TransferBuilder = exports.PolyxBaseBuilder = exports.TransactionBuilderFactory = exports.TransactionBuilder = exports.Transaction = exports.SingletonRegistry = exports.KeyPair = exports.Interface = exports.Errors = exports.Constants = void 0;
var abstract_substrate_1 = require("@bitgo/abstract-substrate");
Object.defineProperty(exports, "Constants", { enumerable: true, get: function () { return abstract_substrate_1.Constants; } });
Object.defineProperty(exports, "Errors", { enumerable: true, get: function () { return abstract_substrate_1.Errors; } });
Object.defineProperty(exports, "Interface", { enumerable: true, get: function () { return abstract_substrate_1.Interface; } });
Object.defineProperty(exports, "KeyPair", { enumerable: true, get: function () { return abstract_substrate_1.KeyPair; } });
Object.defineProperty(exports, "SingletonRegistry", { enumerable: true, get: function () { return abstract_substrate_1.SingletonRegistry; } });
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return abstract_substrate_1.Transaction; } });
Object.defineProperty(exports, "TransactionBuilder", { enumerable: true, get: function () { return abstract_substrate_1.TransactionBuilder; } });
var transactionBuilderFactory_1 = require("./transactionBuilderFactory");
Object.defineProperty(exports, "TransactionBuilderFactory", { enumerable: true, get: function () { return transactionBuilderFactory_1.TransactionBuilderFactory; } });
var baseBuilder_1 = require("./baseBuilder");
Object.defineProperty(exports, "PolyxBaseBuilder", { enumerable: true, get: function () { return baseBuilder_1.PolyxBaseBuilder; } });
var transferBuilder_1 = require("./transferBuilder");
Object.defineProperty(exports, "TransferBuilder", { enumerable: true, get: function () { return transferBuilder_1.TransferBuilder; } });
var registerDidWithCDDBuilder_1 = require("./registerDidWithCDDBuilder");
Object.defineProperty(exports, "RegisterDidWithCDDBuilder", { enumerable: true, get: function () { return registerDidWithCDDBuilder_1.RegisterDidWithCDDBuilder; } });
var preApproveAssetBuilder_1 = require("./preApproveAssetBuilder");
Object.defineProperty(exports, "PreApproveAssetBuilder", { enumerable: true, get: function () { return preApproveAssetBuilder_1.PreApproveAssetBuilder; } });
var tokenTransferBuilder_1 = require("./tokenTransferBuilder");
Object.defineProperty(exports, "TokenTransferBuilder", { enumerable: true, get: function () { return tokenTransferBuilder_1.TokenTransferBuilder; } });
var rejectInstructionBuilder_1 = require("./rejectInstructionBuilder");
Object.defineProperty(exports, "RejectInstructionBuilder", { enumerable: true, get: function () { return rejectInstructionBuilder_1.RejectInstructionBuilder; } });
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "PolyxTransaction", { enumerable: true, get: function () { return transaction_1.Transaction; } });
var bondExtraBuilder_1 = require("./bondExtraBuilder");
Object.defineProperty(exports, "BondExtraBuilder", { enumerable: true, get: function () { return bondExtraBuilder_1.BondExtraBuilder; } });
var batchStakingBuilder_1 = require("./batchStakingBuilder");
Object.defineProperty(exports, "BatchBuilder", { enumerable: true, get: function () { return batchStakingBuilder_1.BatchStakingBuilder; } });
var batchUnstakingBuilder_1 = require("./batchUnstakingBuilder");
Object.defineProperty(exports, "BatchUnstakingBuilder", { enumerable: true, get: function () { return batchUnstakingBuilder_1.BatchUnstakingBuilder; } });
var unbondBuilder_1 = require("./unbondBuilder");
Object.defineProperty(exports, "UnbondBuilder", { enumerable: true, get: function () { return unbondBuilder_1.UnbondBuilder; } });
var withdrawUnbondedBuilder_1 = require("./withdrawUnbondedBuilder");
Object.defineProperty(exports, "WithdrawUnbondedBuilder", { enumerable: true, get: function () { return withdrawUnbondedBuilder_1.WithdrawUnbondedBuilder; } });
const utils_1 = __importDefault(require("./utils"));
var utils_2 = require("./utils");
Object.defineProperty(exports, "Utils", { enumerable: true, get: function () { return utils_2.Utils; } });
Object.defineProperty(exports, "utils", { enumerable: true, get: function () { return __importDefault(utils_2).default; } });
__exportStar(require("./iface"), exports);
/**
 * Checks if a string is a valid Polymesh DID (Decentralized Identifier)
 * DIDs are 32-byte hex strings (0x prefix + 64 hex characters)
 *
 * @param {string} did - The string to validate
 * @returns {boolean} true if valid DID format, false otherwise
 */
const isValidDid = (did) => utils_1.default.isValidDid(did);
exports.isValidDid = isValidDid;
