// @bitgo/sdk-coin-polyx/dist/src/lib/preApproveAssetBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreApproveAssetBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
const baseBuilder_1 = require("./baseBuilder");
const iface_1 = require("./iface");
const txnSchema_1 = require("./txnSchema");
const transaction_1 = require("./transaction");
class PreApproveAssetBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.TrustLine;
    }
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.preApproveAsset({
            assetId: this._assetId,
        }, baseTxInfo);
    }
    /**
     * Sets the asset ID for the pre-approval transaction.
     *
     * @param {string} assetId - The ID of the asset to be pre-approved.
     * @returns {this} The current instance of the builder.
     */
    assetId(assetId) {
        this._assetId = assetId;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name === iface_1.MethodNames.PreApproveAsset) {
            const txMethod = this._method.args;
            this.assetId(txMethod.assetId);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected preApproveAsset`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === iface_1.MethodNames.PreApproveAsset) {
            const txMethod = decodedTxn.method.args;
            const assetId = txMethod.assetId;
            const validationResult = txnSchema_1.PreApproveAssetTransactionSchema.validate({ assetId });
            if (!validationResult) {
                throw new sdk_core_1.InvalidTransactionError('Invalid transaction: assetId is required');
            }
        }
    }
    /**
     * Construct a transaction to pre-approve an asset
     *
     * @param {PreApproveAssetArgs} args Arguments to be passed to the preApproveAsset method
     * @param {Interface.CreateBaseTxInfo} info Base txn info required to construct the pre-approve asset txn
     * @returns {UnsignedTransaction} an unsigned transaction for asset pre-approval
     */
    preApproveAsset(args, info) {
        console.log(`PreApproveAssetBuilder: preApproveAsset called with args: ${JSON.stringify(args)}`);
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'preApproveAsset',
                pallet: 'asset',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.PreApproveAssetBuilder = PreApproveAssetBuilder;
