// @bitgo/sdk-coin-polyx/dist/src/lib/registerDidWithCDDBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterDidWithCDDBuilder = void 0;
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const baseBuilder_1 = require("./baseBuilder");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
const sdk_core_1 = require("@bitgo/sdk-core");
const iface_1 = require("./iface");
const txnSchema_1 = require("./txnSchema");
const transaction_1 = require("./transaction");
class RegisterDidWithCDDBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.WalletInitialization;
    }
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.RegisterDidWithCDD({
            targetAccount: this._to,
            secondaryKeys: [],
            expiry: null,
        }, baseTxInfo);
    }
    /**
     *
     * The destination address for transfer transaction.
     *
     * @param {string} dest
     * @returns {TransferBuilder} This transfer builder.
     */
    to({ address }) {
        this.validateAddress({ address });
        this._to = address;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name === iface_1.MethodNames.RegisterDidWithCDD) {
            const txMethod = this._method.args;
            this.to({ address: abstract_substrate_1.utils.decodeSubstrateAddress(txMethod.targetAccount, this.getAddressFormat()) });
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected RegisterDidWithCDD`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === iface_1.MethodNames.RegisterDidWithCDD) {
            const txMethod = decodedTxn.method.args;
            const targetAccount = txMethod.targetAccount;
            const secondaryKeys = txMethod.secondaryKeys;
            const expiry = txMethod.expiry;
            const validationResult = txnSchema_1.RegisterDidWithCDDTransactionSchema.validate({ targetAccount, secondaryKeys, expiry });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid transaction: ${validationResult.error.message}`);
            }
        }
    }
    /**
     * Construct a transaction to register a DID with CDD information
     *
     * @param {RegisterDidWithCDDArgs} args Arguments to be passed to the cddRegisterDidWithCdd method
     * @param {Interface.CreateBaseTxInfo} info Base txn info required to construct the DID registration txn
     * @returns {UnsignedTransaction} an unsigned transaction for DID registration with CDD
     */
    RegisterDidWithCDD(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'cddRegisterDidWithCdd',
                pallet: 'identity',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.RegisterDidWithCDDBuilder = RegisterDidWithCDDBuilder;
