// @bitgo/sdk-coin-polyx/dist/src/lib/rejectInstructionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RejectInstructionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const baseBuilder_1 = require("./baseBuilder");
const iface_1 = require("./iface");
const transaction_1 = require("./transaction");
const txnSchema_1 = require("./txnSchema");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
class RejectInstructionBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.RejectInstruction;
    }
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.rejectInstruction({
            id: this._instructionId,
            portfolio: {
                did: this._portfolioDID,
                kind: iface_1.PortfolioKind.Default,
            },
            numberOfAssets: {
                fungible: 1,
                nonFungible: 0,
                offChain: 0,
            },
        }, baseTxInfo);
    }
    /**
     * @param instructionId - The ID of the instruction to be rejected
     * @returns {this}
     */
    instructionId(instructionId) {
        this._instructionId = instructionId;
        return this;
    }
    /**
     * @param portfolioDID - The DID of the portfolio associated with the instruction
     * @returns {this}
     */
    portfolioDID(portfolioDID) {
        this._portfolioDID = portfolioDID;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name === iface_1.MethodNames.RejectInstruction) {
            const txMethod = this._method.args;
            this._instructionId = txMethod.id;
            this._portfolioDID = txMethod.portfolio.did;
        }
        else {
            throw new Error(`Cannot build from transaction with method ${this._method?.name} for RejectInstructionBuilder`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === iface_1.MethodNames.RejectInstruction) {
            const txMethod = decodedTxn.method.args;
            const id = txMethod.id;
            const portfolio = txMethod.portfolio;
            const validationResult = txnSchema_1.RejectInstructionTransactionSchema.validate({
                id,
                portfolio,
            });
            if (validationResult.error) {
                throw new Error(`Invalid transaction: ${validationResult.error.message}`);
            }
        }
    }
    rejectInstruction(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'rejectInstruction',
                pallet: 'settlement',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.RejectInstructionBuilder = RejectInstructionBuilder;
