// @bitgo/sdk-coin-polyx/dist/src/lib/tokenTransferBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const baseBuilder_1 = require("./baseBuilder");
const iface_1 = require("./iface");
const transaction_1 = require("./transaction");
const txnSchema_1 = require("./txnSchema");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
class TokenTransferBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.SendToken;
    }
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.addAndAffirmWithMediators({
            venueId: null,
            settlementType: iface_1.SettlementType.SettleOnAffirmation,
            tradeDate: null,
            valueDate: null,
            legs: [
                {
                    fungible: {
                        sender: {
                            did: this._fromDID,
                            kind: iface_1.PortfolioKind.Default,
                        },
                        receiver: {
                            did: this._toDID,
                            kind: iface_1.PortfolioKind.Default,
                        },
                        assetId: this._assetId,
                        amount: this._amount,
                    },
                },
            ],
            portfolios: [
                {
                    did: this._fromDID,
                    kind: iface_1.PortfolioKind.Default,
                },
            ],
            instructionMemo: this._memo,
            mediators: [],
        }, baseTxInfo);
    }
    /**
     * Sets the amount to transfer.
     *
     * @param {string} amount - The amount to transfer.
     * @returns {this} The current instance of the builder.
     */
    assetId(assetId) {
        this._assetId = assetId;
        return this;
    }
    /**
     * Sets the amount to transfer.
     *
     * @param {string} amount - The amount to transfer.
     * @returns {this} The current instance of the builder.
     */
    amount(amount) {
        this._amount = amount;
        return this;
    }
    /**
     * Sets the memo for the transaction.
     * Pads the memo on the left with zeros to ensure it is 32 characters long.
     *
     * @param {string} memo - The memo for the transaction.
     * @returns {this} The current instance of the builder.
     */
    memo(memo) {
        const paddedMemo = memo.padStart(32, '0');
        this._memo = paddedMemo;
        return this;
    }
    /**
     * Sets the sender DID.
     *
     * @param {string} fromDID - The sender DID.
     * @returns {this} The current instance of the builder.
     */
    fromDID(fromDID) {
        this._fromDID = fromDID;
        return this;
    }
    /**
     * Sets the receiver DID.
     *
     * @param {string} toDID - The receiver DID.
     * @returns {this} The current instance of the builder.
     */
    toDID(toDID) {
        this._toDID = toDID;
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method?.name === iface_1.MethodNames.AddAndAffirmWithMediators) {
            const txMethod = this._method.args;
            this.assetId(txMethod.legs[0].fungible.assetId);
            this.amount(txMethod.legs[0].fungible.amount);
            this.memo(txMethod.instructionMemo);
            this.fromDID(txMethod.legs[0].fungible.sender.did);
            this.toDID(txMethod.legs[0].fungible.receiver.did);
        }
        else {
            throw new Error(`Invalid Transaction Type: ${this._method?.name}. Expected AddAndAffirmWithMediators`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === iface_1.MethodNames.AddAndAffirmWithMediators) {
            const txMethod = decodedTxn.method.args;
            const venueId = txMethod.venueId;
            const settlementType = txMethod.settlementType;
            const tradeDate = txMethod.tradeDate;
            const valueDate = txMethod.valueDate;
            const legs = txMethod.legs;
            const portfolios = txMethod.portfolios;
            const instructionMemo = txMethod.instructionMemo;
            const mediators = txMethod.mediators;
            const validationResult = txnSchema_1.AddAndAffirmWithMediatorsTransactionSchema.validate({
                venueId,
                settlementType,
                tradeDate,
                valueDate,
                legs,
                portfolios,
                instructionMemo,
                mediators,
            });
            if (validationResult.error) {
                throw new Error(`Invalid transaction: ${validationResult.error.message}`);
            }
        }
    }
    addAndAffirmWithMediators(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'addAndAffirmWithMediators',
                pallet: 'settlement',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
