// @bitgo/sdk-coin-polyx/dist/src/lib/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const baseBuilder_1 = require("./baseBuilder");
const txwrapper_core_1 = require("@substrate/txwrapper-core");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("./utils"));
class TransferBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.material(utils_1.default.getMaterial(_coinConfig.network.type));
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    /**
     * Construct an unsigned `transferWithMemo` transaction using the provided details.
     *
     * @returns {UnsignedTransaction} The constructed unsigned transferWithMemo transaction.
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return this.TransferWithMemo({
            dest: { id: this._to },
            value: this._amount,
            memo: this._memo,
        }, baseTxInfo);
    }
    /**
     *
     * The amount for transfer transaction.
     *
     * @param {string} amount
     * @returns {TransferBuilder} This transfer builder.
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     *
     * The destination address for transfer transaction.
     *
     * @param {string} dest
     * @returns {TransferBuilder} This transfer builder.
     */
    to({ address }) {
        this.validateAddress({ address });
        this._to = address;
        return this;
    }
    /**
     * The memo to attach to the transfer transaction.
     * Pads the memo on the left with zeros to ensure it is 32 characters long.
     *
     * @param {string} memo The memo string to include.
     * @returns {TransferBuilder} This transfer builder.
     */
    memo(memo) {
        const paddedMemo = memo.padStart(32, '0');
        this._memo = paddedMemo;
        return this;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn, rawTransaction) {
        if (decodedTxn.method?.name === abstract_substrate_1.Interface.MethodNames.TransferWithMemo) {
            const txMethod = decodedTxn.method.args;
            const amount = `${txMethod.value}`;
            const to = txMethod.dest.id;
            const memo = txMethod.memo;
            const validationResult = abstract_substrate_1.Schema.TransferWithMemoTransactionSchema.validate({ amount, to, memo });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid transaction: ${validationResult.error.message}`);
            }
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (!this._method || !this._method.args) {
            throw new sdk_core_1.InvalidTransactionError('Transaction method or args are undefined');
        }
        if (this._method?.name === abstract_substrate_1.Interface.MethodNames.TransferWithMemo) {
            const txMethod = this._method.args;
            this.amount(txMethod.value);
            this.to({
                address: utils_1.default.decodeSubstrateAddress(txMethod.dest.id, utils_1.default.getAddressFormat(this._coinConfig.name)),
            });
            this.memo(txMethod.memo);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method.name}. Expected transferWithMemo`);
        }
        return tx;
    }
    /**
     * Construct a transaction to transfer funds with an attached memo.
     *
     * @param {Interface.TransferWithMemoArgs} args Arguments to be passed to the transferWithMemo method
     * @param {Interface.CreateBaseTxInfo} info Base txn info required to construct the transfer transaction
     * @returns {UnsignedTransaction} An unsigned transferWithMemo transaction
     */
    TransferWithMemo(args, info) {
        return (0, txwrapper_core_1.defineMethod)({
            method: {
                args,
                name: 'transferWithMemo',
                pallet: 'balances',
            },
            ...info.baseTxInfo,
        }, info.options);
    }
}
exports.TransferBuilder = TransferBuilder;
