// @bitgo/sdk-coin-polyx/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const transferBuilder_1 = require("./transferBuilder");
const registerDidWithCDDBuilder_1 = require("./registerDidWithCDDBuilder");
const bondExtraBuilder_1 = require("./bondExtraBuilder");
const batchStakingBuilder_1 = require("./batchStakingBuilder");
const batchUnstakingBuilder_1 = require("./batchUnstakingBuilder");
const unbondBuilder_1 = require("./unbondBuilder");
const withdrawUnbondedBuilder_1 = require("./withdrawUnbondedBuilder");
const utils_1 = __importDefault(require("./utils"));
const _1 = require("./");
const iface_1 = require("./iface");
const preApproveAssetBuilder_1 = require("./preApproveAssetBuilder");
const tokenTransferBuilder_1 = require("./tokenTransferBuilder");
const rejectInstructionBuilder_1 = require("./rejectInstructionBuilder");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._material = utils_1.default.getMaterial(_coinConfig.network.type);
    }
    getTransferBuilder() {
        return new transferBuilder_1.TransferBuilder(this._coinConfig).material(this._material);
    }
    getRegisterDidWithCDDBuilder() {
        return new registerDidWithCDDBuilder_1.RegisterDidWithCDDBuilder(this._coinConfig).material(this._material);
    }
    getPreApproveAssetBuilder() {
        return new preApproveAssetBuilder_1.PreApproveAssetBuilder(this._coinConfig).material(this._material);
    }
    getTokenTransferBuilder() {
        return new tokenTransferBuilder_1.TokenTransferBuilder(this._coinConfig).material(this._material);
    }
    getRejectInstructionBuilder() {
        return new rejectInstructionBuilder_1.RejectInstructionBuilder(this._coinConfig).material(this._material);
    }
    getBondExtraBuilder() {
        return new bondExtraBuilder_1.BondExtraBuilder(this._coinConfig).material(this._material);
    }
    getBatchBuilder() {
        return new batchStakingBuilder_1.BatchStakingBuilder(this._coinConfig).material(this._material);
    }
    getBatchUnstakingBuilder() {
        return new batchUnstakingBuilder_1.BatchUnstakingBuilder(this._coinConfig).material(this._material);
    }
    getUnbondBuilder() {
        return new unbondBuilder_1.UnbondBuilder(this._coinConfig).material(this._material);
    }
    getWithdrawUnbondedBuilder() {
        return new withdrawUnbondedBuilder_1.WithdrawUnbondedBuilder(this._coinConfig).material(this._material);
    }
    getWalletInitializationBuilder() {
        throw new sdk_core_1.NotImplementedError(`walletInitialization for ${this._coinConfig.name} not implemented`);
    }
    from(rawTxn) {
        const builder = this.getBuilder(rawTxn);
        builder.from(rawTxn);
        return builder;
    }
    material(material) {
        this._material = material;
        return this;
    }
    getBuilder(rawTxn) {
        const registry = _1.SingletonRegistry.getInstance(this._material);
        const decodedTxn = (0, txwrapper_polkadot_1.decode)(rawTxn, {
            metadataRpc: this._material.metadata,
            registry: registry,
        });
        const methodName = decodedTxn.method?.name;
        if (methodName === _1.Interface.MethodNames.TransferWithMemo) {
            return this.getTransferBuilder();
        }
        else if (methodName === iface_1.MethodNames.RegisterDidWithCDD) {
            return this.getRegisterDidWithCDDBuilder();
        }
        else if (methodName === iface_1.MethodNames.PreApproveAsset) {
            return this.getPreApproveAssetBuilder();
        }
        else if (methodName === iface_1.MethodNames.AddAndAffirmWithMediators) {
            return this.getTokenTransferBuilder();
        }
        else if (methodName === iface_1.MethodNames.RejectInstruction) {
            return this.getRejectInstructionBuilder();
        }
        else if (methodName === 'bondExtra') {
            return this.getBondExtraBuilder();
        }
        else if (methodName === 'batchAll') {
            const args = decodedTxn.method.args;
            if (args.calls && args.calls.length === 2) {
                // Decode method names from the calls using utils.decodeMethodName
                const firstCallMethod = utils_1.default.decodeMethodName(args.calls[0], registry);
                const secondCallMethod = utils_1.default.decodeMethodName(args.calls[1], registry);
                // Check for batch staking pattern: bond + nominate
                if (firstCallMethod === 'bond' && secondCallMethod === 'nominate') {
                    return this.getBatchBuilder();
                }
                // Check for batch unstaking pattern: chill + unbond
                if (firstCallMethod === 'chill' && secondCallMethod === 'unbond') {
                    return this.getBatchUnstakingBuilder();
                }
            }
            // Fall back to general batch builder for other batchAll cases
            return this.getBatchBuilder();
        }
        else if (methodName === 'batch') {
            return this.getBatchBuilder();
        }
        else if (methodName === 'bond') {
            return this.getBatchBuilder();
        }
        else if (methodName === 'nominate') {
            return this.getBatchBuilder();
        }
        else if (methodName === 'unbond') {
            return this.getUnbondBuilder();
        }
        else if (methodName === 'withdrawUnbonded') {
            return this.getWithdrawUnbondedBuilder();
        }
        throw new Error('Transaction cannot be parsed or has an unsupported transaction type');
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
