// @bitgo/sdk-coin-polyx/dist/src/lib/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const abstract_substrate_1 = require("@bitgo/abstract-substrate");
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const keyring_1 = require("@polkadot/keyring");
const statics_1 = require("@bitgo/statics");
const utils_1 = __importDefault(require("./utils"));
class Transaction extends abstract_substrate_1.Transaction {
    /**
     * Override the getAddressFormat method to return different values based on network type
     * Returns 12 for mainnet and 42 for testnet
     *
     * @returns {number} The address format to use
     * @override
     */
    getAddressFormat() {
        return utils_1.default.getAddressFormat(this._coinConfig.name);
    }
    /**
     * Helper method to find a token name by assetId
     * @param assetId The assetId to search for
     * @returns The token name if found
     * @throws {InvalidTransactionError} If no matching token is found
     */
    getTokenNameByAssetId(assetId) {
        // Search through all coins to find the one with matching assetId
        let foundToken;
        statics_1.coins.forEach((coin) => {
            if (coin instanceof statics_1.PolyxCoin && coin.assetId === assetId) {
                foundToken = coin.name;
            }
        });
        if (!foundToken) {
            throw new sdk_core_1.InvalidTransactionError(`No token found for assetId: ${assetId}`);
        }
        return foundToken;
    }
    /** @inheritdoc */
    toJson() {
        if (!this._substrateTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        const decodedTx = (0, txwrapper_polkadot_1.decode)(this._substrateTransaction, {
            metadataRpc: this._substrateTransaction.metadataRpc,
            registry: this._registry,
            isImmortalEra: abstract_substrate_1.utils.isZeroHex(this._substrateTransaction.era),
        });
        const result = {
            id: txwrapper_polkadot_1.construct.txHash(this.toBroadcastFormat()),
            sender: decodedTx.address,
            referenceBlock: decodedTx.blockHash,
            blockNumber: decodedTx.blockNumber,
            genesisHash: decodedTx.genesisHash,
            nonce: decodedTx.nonce,
            specVersion: decodedTx.specVersion,
            transactionVersion: decodedTx.transactionVersion,
            eraPeriod: decodedTx.eraPeriod,
            chainName: this._chainName,
            tip: decodedTx.tip ? Number(decodedTx.tip) : 0,
        };
        const txMethod = decodedTx.method.args;
        if (this.type === sdk_core_1.TransactionType.WalletInitialization) {
            const { targetAccount } = txMethod;
            const keypairDest = new abstract_substrate_1.KeyPair({
                pub: Buffer.from((0, keyring_1.decodeAddress)(targetAccount)).toString('hex'),
            });
            result.to = keypairDest.getAddress(this.getAddressFormat());
            result.amount = '0'; // RegisterDidWithCDD does not transfer any value
        }
        else if (this.type === sdk_core_1.TransactionType.TrustLine) {
            const { assetId } = txMethod;
            result.assetId = assetId;
            result.sender = decodedTx.address;
            result.amount = '0'; // Pre-approval does not transfer any value
        }
        else if (this.type === sdk_core_1.TransactionType.SendToken) {
            const sendTokenArgs = txMethod;
            result.fromDID = sendTokenArgs.legs[0].fungible.sender.did;
            result.toDID = sendTokenArgs.legs[0].fungible.receiver.did;
            result.amount = sendTokenArgs.legs[0].fungible.amount.toString();
            result.assetId = sendTokenArgs.legs[0].fungible.assetId;
            result.memo = sendTokenArgs.instructionMemo;
        }
        else if (this.type === sdk_core_1.TransactionType.RejectInstruction) {
            const rejectInstructionArgs = txMethod;
            result.instructionId = rejectInstructionArgs.id;
            result.portfolioDID = rejectInstructionArgs.portfolio.did;
            result.amount = '0'; // Reject instruction does not transfer any value
        }
        else {
            return super.toJson();
        }
        return result;
    }
    /**
     * Load the input and output data on this transaction.
     */
    loadInputsAndOutputs() {
        super.loadInputsAndOutputs();
        const decodedTx = (0, txwrapper_polkadot_1.decode)(this._substrateTransaction, {
            metadataRpc: this._substrateTransaction.metadataRpc,
            registry: this._registry,
            isImmortalEra: abstract_substrate_1.utils.isZeroHex(this._substrateTransaction.era),
        });
        if (this.type === sdk_core_1.TransactionType.WalletInitialization) {
            this.decodeInputsAndOutputsForRegisterDidWithCDD(decodedTx);
        }
        else if (this.type === sdk_core_1.TransactionType.TrustLine) {
            this.decodeInputsAndOutputsForPreApproveAsset(decodedTx);
        }
        else if (this.type === sdk_core_1.TransactionType.SendToken) {
            this.decodeInputsAndOutputsForSendToken(decodedTx);
        }
        else if (this.type === sdk_core_1.TransactionType.RejectInstruction) {
            this.decodeInputsAndOutputsForRejectInstruction(decodedTx);
        }
    }
    decodeInputsAndOutputsForRegisterDidWithCDD(decodedTx) {
        const txMethod = decodedTx.method.args;
        const keypairDest = new abstract_substrate_1.KeyPair({
            pub: Buffer.from((0, keyring_1.decodeAddress)(txMethod.targetAccount)).toString('hex'),
        });
        const to = keypairDest.getAddress(this.getAddressFormat());
        const value = '0';
        const from = decodedTx.address;
        this._inputs.push({
            address: from,
            value,
            coin: this._coinConfig.name,
        });
        this._outputs.push({
            address: to,
            value,
            coin: this._coinConfig.name,
        });
    }
    decodeInputsAndOutputsForPreApproveAsset(decodedTx) {
        const txMethod = decodedTx.method.args;
        const sender = decodedTx.address;
        const value = '0'; // Pre-approval does not transfer any value
        const tokenName = this.getTokenNameByAssetId(txMethod.assetId);
        this._inputs.push({
            address: sender,
            value,
            coin: tokenName,
        });
        this._outputs.push({
            address: sender, // In pre-approval, the output is the same as the input
            value,
            coin: tokenName,
        });
    }
    decodeInputsAndOutputsForSendToken(decodedTx) {
        const txMethod = decodedTx.method.args;
        const fromDID = txMethod.legs[0].fungible.sender.did;
        const toDID = txMethod.legs[0].fungible.receiver.did;
        const amount = txMethod.legs[0].fungible.amount.toString();
        const assetId = txMethod.legs[0].fungible.assetId;
        const tokenName = this.getTokenNameByAssetId(assetId);
        this._inputs.push({
            address: fromDID,
            value: amount,
            coin: tokenName,
        });
        this._outputs.push({
            address: toDID,
            value: amount,
            coin: tokenName,
        });
    }
    decodeInputsAndOutputsForRejectInstruction(decodedTx) {
        const txMethod = decodedTx.method.args;
        const portfolioDID = txMethod.portfolio.did;
        const value = '0'; // Reject instruction does not transfer any value
        this._inputs.push({
            address: portfolioDID,
            value,
            coin: this._coinConfig.name,
        });
        this._outputs.push({
            address: portfolioDID,
            value,
            coin: this._coinConfig.name,
        });
    }
}
exports.Transaction = Transaction;
