// @bitgo/sdk-coin-polyx/dist/src/lib/txnSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawUnbondedTransactionSchema = exports.UnbondTransactionSchema = exports.BatchUnstakingTransactionSchema = exports.bondSchema = exports.BatchTransactionSchema = exports.NominateTransactionSchema = exports.BondExtraTransactionSchema = exports.RejectInstructionTransactionSchema = exports.AddAndAffirmWithMediatorsTransactionSchema = exports.PreApproveAssetTransactionSchema = exports.RegisterDidWithCDDTransactionSchema = void 0;
const joi_1 = __importDefault(require("joi"));
const utils_1 = __importDefault(require("./utils"));
const addressSchema = joi_1.default.string().custom((addr) => utils_1.default.isValidAddress(addr));
exports.RegisterDidWithCDDTransactionSchema = joi_1.default.object({
    targetAccount: addressSchema.required(),
    secondaryKeys: joi_1.default.array().length(0).required(),
    expiry: joi_1.default.valid(null).required(),
});
exports.PreApproveAssetTransactionSchema = joi_1.default.object({
    assetId: joi_1.default.string().required(),
});
exports.AddAndAffirmWithMediatorsTransactionSchema = joi_1.default.object({
    venueId: joi_1.default.valid(null).required(),
    settlementType: joi_1.default
        .object({
        settleOnAffirmation: joi_1.default.valid(null),
    })
        .required(),
    tradeDate: joi_1.default.valid(null).required(),
    valueDate: joi_1.default.valid(null).required(),
    legs: joi_1.default
        .array()
        .items(joi_1.default.object({
        fungible: joi_1.default
            .object({
            sender: joi_1.default
                .object({
                did: addressSchema.required(),
                kind: joi_1.default
                    .object({
                    default: joi_1.default.valid(null),
                })
                    .required(),
            })
                .required(),
            receiver: joi_1.default
                .object({
                did: addressSchema.required(),
                kind: joi_1.default
                    .object({
                    default: joi_1.default.valid(null),
                })
                    .required(),
            })
                .required(),
            assetId: joi_1.default.string().required(),
            amount: joi_1.default.number().required(),
        })
            .required(),
    }))
        .required(),
    portfolios: joi_1.default
        .array()
        .items(joi_1.default.object({
        did: addressSchema.required(),
        kind: joi_1.default
            .object({
            default: joi_1.default.valid(null),
        })
            .required(),
    }))
        .required(),
    instructionMemo: joi_1.default.string().required(),
    mediators: joi_1.default.array().length(0).required(),
});
exports.RejectInstructionTransactionSchema = joi_1.default.object({
    id: joi_1.default.string().required(),
    portfolio: joi_1.default
        .object({
        did: addressSchema.required(),
        kind: joi_1.default
            .object({
            default: joi_1.default.valid(null),
        })
            .required(),
    })
        .required(),
    numberOfAssets: joi_1.default
        .object({
        fungible: joi_1.default.number().required(),
        nonFungible: joi_1.default.number().required(),
        offChain: joi_1.default.number().required(),
    })
        .optional(),
});
// For standalone bondExtra transactions
exports.BondExtraTransactionSchema = joi_1.default.object({
    value: joi_1.default.string().required(),
});
// For nominate transactions in batch
exports.NominateTransactionSchema = joi_1.default.object({
    validators: joi_1.default.array().items(addressSchema).min(1).required(),
});
// For batch validation
exports.BatchTransactionSchema = {
    validate: (value) => joi_1.default
        .object({
        amount: joi_1.default.string().required(),
        controller: addressSchema.required(),
        // Payee can be a string or an object with Account property
        payee: joi_1.default
            .alternatives()
            .try(joi_1.default.string(), joi_1.default.object({
            Account: addressSchema,
        }))
            .required(),
        validators: joi_1.default.array().items(addressSchema).min(1).required(),
    })
        .validate(value),
    validateBond: (value) => joi_1.default
        .object({
        value: joi_1.default.string().required(),
        controller: addressSchema.required(),
        payee: joi_1.default
            .alternatives()
            .try(joi_1.default.string(), joi_1.default.object({
            Account: addressSchema,
        }))
            .required(),
    })
        .validate(value),
    validateNominate: (value) => joi_1.default
        .object({
        validators: joi_1.default.array().items(addressSchema).min(1).required(),
    })
        .validate(value),
};
exports.bondSchema = joi_1.default.object({
    value: joi_1.default.string().required(),
    controller: addressSchema.required(),
    payee: joi_1.default
        .alternatives()
        .try(joi_1.default.string().valid('Staked', 'Stash', 'Controller'), joi_1.default.object({
        Account: addressSchema.required(),
    }))
        .required(),
});
exports.BatchUnstakingTransactionSchema = {
    validate: (value) => joi_1.default
        .object({
        value: joi_1.default.string().required(),
    })
        .validate(value),
};
exports.UnbondTransactionSchema = {
    validate: (value) => joi_1.default
        .object({
        value: joi_1.default.string().required(),
    })
        .validate(value),
};
exports.WithdrawUnbondedTransactionSchema = {
    validate: (value) => joi_1.default
        .object({
        slashingSpans: joi_1.default.number().min(0).required(),
    })
        .validate(value),
};
