// @bitgo/sdk-coin-polyx/dist/src/lib/unbondBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnbondBuilder = void 0;
const baseBuilder_1 = require("./baseBuilder");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const sdk_core_1 = require("@bitgo/sdk-core");
const txnSchema_1 = require("./txnSchema");
const utils_1 = __importDefault(require("./utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class UnbondBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.material(utils_1.default.getMaterial(_coinConfig.network.type));
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingDeactivate;
    }
    /**
     * Build the unbond transaction
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return txwrapper_polkadot_1.methods.staking.unbond({
            value: this._amount,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    }
    /**
     * Set the amount to unbond
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    }
    /**
     * Get the amount to unbond
     */
    getAmount() {
        return this._amount;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn) {
        const methodName = decodedTxn.method?.name;
        if (methodName === 'unbond') {
            const txMethod = decodedTxn.method.args;
            const value = txMethod.value;
            const validationResult = txnSchema_1.UnbondTransactionSchema.validate({
                value,
            });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`Invalid unbond transaction: ${validationResult.error.message}`);
            }
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction type: ${methodName}. Expected unbond`);
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        const methodName = this._method?.name;
        if (methodName === 'unbond' && this._method) {
            const txMethod = this._method.args;
            this.amount(txMethod.value);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${methodName}. Expected unbond`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(tx) {
        super.validateTransaction(tx);
        this.validateFields();
    }
    /**
     * Validate the unbond fields
     */
    validateFields() {
        const validationResult = txnSchema_1.UnbondTransactionSchema.validate({
            value: this._amount,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid unbond transaction: ${validationResult.error.message}`);
        }
    }
    /**
     * Validates fields for testing
     */
    testValidateFields() {
        this.validateFields();
    }
}
exports.UnbondBuilder = UnbondBuilder;
