// @bitgo/sdk-coin-polyx/dist/src/lib/withdrawUnbondedBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawUnbondedBuilder = void 0;
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const baseBuilder_1 = require("./baseBuilder");
const txnSchema_1 = require("./txnSchema");
class WithdrawUnbondedBuilder extends baseBuilder_1.PolyxBaseBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._slashingSpans = 0;
    }
    /**
     * Withdraw unbonded tokens after the unbonding period has passed
     *
     * @returns {UnsignedTransaction} an unsigned Polyx transaction
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        return txwrapper_polkadot_1.methods.staking.withdrawUnbonded({
            numSlashingSpans: this._slashingSpans,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingWithdraw;
    }
    /**
     * The number of slashing spans, typically 0 for most users
     *
     * @param {number} slashingSpans
     * @returns {WithdrawUnbondedBuilder} This withdrawUnbonded builder.
     */
    slashingSpans(slashingSpans) {
        this.validateValue(new bignumber_js_1.default(slashingSpans));
        this._slashingSpans = slashingSpans;
        return this;
    }
    /**
     * Get the slashing spans
     */
    getSlashingSpans() {
        return this._slashingSpans;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn) {
        if (decodedTxn.method?.name === 'withdrawUnbonded') {
            const txMethod = decodedTxn.method.args;
            const slashingSpans = txMethod.numSlashingSpans;
            const validationResult = txnSchema_1.WithdrawUnbondedTransactionSchema.validate({ slashingSpans });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`WithdrawUnbonded Transaction validation failed: ${validationResult.error.message}`);
            }
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction type: ${decodedTxn.method?.name}. Expected withdrawUnbonded`);
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        if (this._method && this._method.name === 'withdrawUnbonded') {
            const txMethod = this._method.args;
            this.slashingSpans(txMethod.numSlashingSpans);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${this._method?.name}. Expected withdrawUnbonded`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._slashingSpans);
    }
    validateFields(slashingSpans) {
        const validationResult = txnSchema_1.WithdrawUnbondedTransactionSchema.validate({
            slashingSpans,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`WithdrawUnbonded Builder Transaction validation failed: ${validationResult.error.message}`);
        }
    }
}
exports.WithdrawUnbondedBuilder = WithdrawUnbondedBuilder;
