// @bitgo/sdk-coin-dot/dist/src/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.Interface = exports.RemoveProxyBuilder = exports.NativeTransferBuilder = exports.SingletonRegistry = exports.TransactionBuilderFactory = exports.ClaimBuilder = exports.WithdrawUnstakedBuilder = exports.UnnominateBuilder = exports.UnstakeBuilder = exports.BatchTransactionBuilder = exports.AddressInitializationBuilder = exports.TransferBuilder = exports.StakingBuilder = exports.TransactionBuilder = exports.Transaction = exports.KeyPair = void 0;
const Utils = __importStar(require("./utils"));
exports.Utils = Utils;
const Interface = __importStar(require("./iface"));
exports.Interface = Interface;
var keyPair_1 = require("./keyPair");
Object.defineProperty(exports, "KeyPair", { enumerable: true, get: function () { return keyPair_1.KeyPair; } });
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return transaction_1.Transaction; } });
var transactionBuilder_1 = require("./transactionBuilder");
Object.defineProperty(exports, "TransactionBuilder", { enumerable: true, get: function () { return transactionBuilder_1.TransactionBuilder; } });
var stakingBuilder_1 = require("./stakingBuilder");
Object.defineProperty(exports, "StakingBuilder", { enumerable: true, get: function () { return stakingBuilder_1.StakingBuilder; } });
var transferBuilder_1 = require("./transferBuilder");
Object.defineProperty(exports, "TransferBuilder", { enumerable: true, get: function () { return transferBuilder_1.TransferBuilder; } });
var addressInitializationBuilder_1 = require("./addressInitializationBuilder");
Object.defineProperty(exports, "AddressInitializationBuilder", { enumerable: true, get: function () { return addressInitializationBuilder_1.AddressInitializationBuilder; } });
var batchTransactionBuilder_1 = require("./batchTransactionBuilder");
Object.defineProperty(exports, "BatchTransactionBuilder", { enumerable: true, get: function () { return batchTransactionBuilder_1.BatchTransactionBuilder; } });
var unstakeBuilder_1 = require("./unstakeBuilder");
Object.defineProperty(exports, "UnstakeBuilder", { enumerable: true, get: function () { return unstakeBuilder_1.UnstakeBuilder; } });
var unnominateBuilder_1 = require("./unnominateBuilder");
Object.defineProperty(exports, "UnnominateBuilder", { enumerable: true, get: function () { return unnominateBuilder_1.UnnominateBuilder; } });
var withdrawUnstakedBuilder_1 = require("./withdrawUnstakedBuilder");
Object.defineProperty(exports, "WithdrawUnstakedBuilder", { enumerable: true, get: function () { return withdrawUnstakedBuilder_1.WithdrawUnstakedBuilder; } });
var claimBuilder_1 = require("./claimBuilder");
Object.defineProperty(exports, "ClaimBuilder", { enumerable: true, get: function () { return claimBuilder_1.ClaimBuilder; } });
var transactionBuilderFactory_1 = require("./transactionBuilderFactory");
Object.defineProperty(exports, "TransactionBuilderFactory", { enumerable: true, get: function () { return transactionBuilderFactory_1.TransactionBuilderFactory; } });
var singletonRegistry_1 = require("./singletonRegistry");
Object.defineProperty(exports, "SingletonRegistry", { enumerable: true, get: function () { return singletonRegistry_1.SingletonRegistry; } });
var nativeTransferBuilder_1 = require("./nativeTransferBuilder");
Object.defineProperty(exports, "NativeTransferBuilder", { enumerable: true, get: function () { return nativeTransferBuilder_1.NativeTransferBuilder; } });
var proxyBuilder_1 = require("./proxyBuilder");
Object.defineProperty(exports, "RemoveProxyBuilder", { enumerable: true, get: function () { return proxyBuilder_1.RemoveProxyBuilder; } });
