// @bitgo/sdk-core/dist/src/bitgo/utils/tss/baseTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EncryptedSignerShareType = exports.CommitmentType = exports.SignatureShareType = exports.RequestType = exports.TokenType = exports.MPCType = exports.ShareType = void 0;
var ShareType;
(function (ShareType) {
    ShareType["R"] = "R";
    ShareType["Commitment"] = "commitment";
    ShareType["G"] = "G";
    ShareType["S"] = "S";
    ShareType["K"] = "K";
    ShareType["MuDelta"] = "MuDelta";
    ShareType["PaillierModulus"] = "PaillierModulus";
    ShareType["MPCv2Round1"] = "MPCv2Round1";
    ShareType["MPCv2Round2"] = "MPCv2Round2";
    ShareType["MPCv2Round3"] = "MPCv2Round3";
})(ShareType || (exports.ShareType = ShareType = {}));
var MPCType;
(function (MPCType) {
    MPCType["EDDSA"] = "eddsa";
    MPCType["ECDSA"] = "ecdsa";
})(MPCType || (exports.MPCType = MPCType = {}));
var TokenType;
(function (TokenType) {
    TokenType["ERC721"] = "ERC721";
    TokenType["ERC1155"] = "ERC1155";
    TokenType["ERC20"] = "ERC20";
    TokenType["DIGITAL_ASSET"] = "Digital Asset";
})(TokenType || (exports.TokenType = TokenType = {}));
var RequestType;
(function (RequestType) {
    RequestType[RequestType["tx"] = 0] = "tx";
    RequestType[RequestType["message"] = 1] = "message";
})(RequestType || (exports.RequestType = RequestType = {}));
exports.SignatureShareType = {
    USER: 'user',
    BACKUP: 'backup',
    BITGO: 'bitgo',
};
exports.CommitmentType = {
    COMMITMENT: 'commitment',
    DECOMMITMENT: 'decommitment',
};
exports.EncryptedSignerShareType = {
    ENCRYPTED_SIGNER_SHARE: 'encryptedSignerShare',
    ENCRYPTED_R_SHARE: 'encryptedRShare',
};
