// @bitgo/public-types/src/schema/transactionRequest/intents/baseIntent.ts
import * as t from "io-ts";
import { IntentType } from "./intentType";
import { Optional, optionalString } from "../../../utils";
import { RecipientEntry } from "./recipientEntry";
import { TokenRecipientEntry } from "./tokenRecipientEntry";
import { Amount } from "./amount";

/**
 * @title Intent
 */
export const BaseIntent = t.intersection([
  t.type({
    intentType: IntentType,
  }),
  t.partial({
    sequenceId: optionalString,
    comment: optionalString,
    nonce: Optional(t.union([t.string, t.number])),
  }),
]);

export type BaseIntent = t.TypeOf<typeof BaseIntent>;

export const BaseIntentWithoutNonce = t.intersection([
  t.type({
    intentType: IntentType,
  }),
  t.partial({
    sequenceId: optionalString,
    comment: optionalString,
  }),
]);

export type BaseIntentWithoutNonce = t.TypeOf<typeof BaseIntentWithoutNonce>;

/**
 * ============================================================================
 * REUSABLE SCHEMA FRAGMENTS FOR ASSET-CONTAINING INTENTS
 * ============================================================================
 *
 * These schemas define reusable building blocks that concrete intent schemas
 * can compose with. They serve multiple purposes:
 *
 * 1. **Schema Reuse**: Concrete intents can intersect with these fragments
 * 2. **Pattern Matching**: getAssetsFromIntent utility matches against these
 * 3. **Type Safety**: Ensures consistency across all asset-containing intents
 * 4. **Documentation**: Clearly shows which intents handle which asset types
 *
 * By using the actual RecipientEntry and TokenRecipientEntry types, we ensure
 * these patterns exactly match the structures used in concrete intents.
 */

/**
 * Schema fragment for intents with standard recipients (payment-like intents)
 *
 * This fragment defines a recipients array using RecipientEntry, which contains
 * address and amount (with value and symbol) fields.
 *
 * Used by:
 * - PaymentIntent (and all variants: AdaPaymentIntent, EthPaymentIntent, etc.)
 * - ConsolidateIntent variants (when they include recipients)
 *
 * Asset extraction: Extracts `symbol` from each recipient's amount
 */
export const BaseIntentWithRecipients = t.type({
  recipients: t.array(RecipientEntry),
});

export type BaseIntentWithRecipients = t.TypeOf<
  typeof BaseIntentWithRecipients
>;

/**
 * Schema fragment for intents with token recipients
 *
 * This fragment defines a recipients array using TokenRecipientEntry, which
 * extends RecipientEntry with optional tokenData containing token-specific info.
 *
 * Used by:
 * - TransferTokenIntent (and variants: EthTransferTokenIntent, etc.)
 * - ConsolidateTokenIntent (and variants: EthConsolidateTokenIntent, etc.)
 *
 * Asset extraction: Extracts tokenName only (the actual asset identifier)
 * Note: tokenType, tokenContractAddress, tokenId are technical fields, not assets
 * Note: This also matches BaseIntentWithRecipients since TokenRecipientEntry extends RecipientEntry
 */
export const BaseIntentWithTokenRecipients = t.type({
  recipients: t.array(TokenRecipientEntry),
});

export type BaseIntentWithTokenRecipients = t.TypeOf<
  typeof BaseIntentWithTokenRecipients
>;

/**
 * Pattern for matching token data in recipients
 * Used to extract tokenName (the asset identifier) from token-related intents
 */
export const BaseIntentWithTokenData = t.type({
  recipients: t.array(
    t.partial({
      tokenData: t.partial({
        tokenName: t.string,
      }),
    }),
  ),
});

export type BaseIntentWithTokenData = t.TypeOf<typeof BaseIntentWithTokenData>;

/**
 * Schema fragment for intents with direct tokenName field
 *
 * This fragment defines a tokenName field directly on the intent object
 * (not nested in recipients).
 *
 * Used by:
 * - TokenApprovalIntent
 *
 * Asset extraction: Extracts tokenName
 */
export const BaseIntentWithTokenName = t.type({
  tokenName: t.string,
});

export type BaseIntentWithTokenName = t.TypeOf<typeof BaseIntentWithTokenName>;

/**
 * Schema fragment for intents with direct amount field
 *
 * This fragment defines an amount field directly on the intent object
 * (not nested in recipients). Common in staking, claiming, and delegation intents.
 *
 * Used by:
 * - StakeIntent variants (AtomStakeIntent, SolStakeIntent, etc.)
 * - UnstakeIntent variants (AtomUnstakeIntent, etc.)
 * - DelegateIntent variants (BscDelegateIntent, TonDelegateIntent, etc.)
 * - ClaimRewardsIntent variants (AdaStakeClaimRewardsIntent, etc.)
 * - WithdrawIntent variants (AtomWithdrawIntent, CosmosWithdrawIntent, etc.)
 *
 * Asset extraction: Extracts symbol from the amount field
 */
export const BaseIntentWithAmount = t.type({
  amount: Amount,
});

export type BaseIntentWithAmount = t.TypeOf<typeof BaseIntentWithAmount>;
