// @bitgo/public-types/src/schema/transactionRequest/intents/cantonTransferAcknowledgeIntent.ts
import * as t from "io-ts";
import { NumberFromString } from "io-ts-types";
import { BaseIntent } from "./baseIntent";
import { intentTypes } from "./intentType";

/**
 * @title Canton Transfer Acknowledge Intent
 *
 * Intent for auto creating a dummy txn for the user to either accept or reject
 * 2-step transfer in `Canton` chain means that the receiver has to explicitly accept/reject the deposit
 * by sending out an acceptance/rejection transaction on-chain so when there is a deposit, we automatically create a
 * txRequest with the details needed for user to take action
 *
 * @param {String} intentType - The type of intent (always set to transferAcknowledge)
 * @param {String} contractId - The contract id that needs to be acted on by the receiver
 * @param {String} senderPartyId - The address (partyId) of the sender
 * @param {String} updateId - The transaction id of the deposit
 * @param {String} amount - The deposit amount (numeric string, decoded to number)
 * @param {Number} expiry - The expiry epoch
 * @param {String} [memoId] - The optional memoId
 * @param {String} [token] - The optional token
 */
export const CantonTransferAcknowledgeIntent = t.intersection([
  BaseIntent,
  t.type({
    intentType: intentTypes.transferAcknowledge,
    contractId: t.string,
    senderPartyId: t.string,
    updateId: t.string,
    amount: NumberFromString,
    expiry: t.number,
  }),
  t.partial({
    memoId: t.string,
    token: t.string,
  }),
]);

export type CantonTransferAcknowledgeIntent = t.TypeOf<
  typeof CantonTransferAcknowledgeIntent
>;
