// @bitgo/public-types/src/schema/transactionRequest/intents/ethBuildOptions.ts
import * as t from "io-ts";
import { AccountBaseBuildOptions } from "./accountBaseBuildOptions";
import { Optional, optionalNumber, optionalString } from "../../../utils";
import { FeeOption } from "./feeOption";
import { EIP1559FeeOptions } from "./eip1559FeeOptions";
import { HopParams } from "./hopParams";

export const EthBuildOptions = t.intersection([
  AccountBaseBuildOptions,
  t.partial({
    feeOptions: Optional(t.union([FeeOption, EIP1559FeeOptions])),
    hopParams: Optional(HopParams),
    receiveAddress: optionalString,
    senderAddressIndex: optionalNumber,
  }),
]);

export type EthBuildOptions = t.TypeOf<typeof EthBuildOptions>;

export const EthBuildOptionsWithoutFeeOptions = t.intersection([
  AccountBaseBuildOptions,
  t.partial({
    hopParams: Optional(HopParams),
    receiveAddress: optionalString,
    senderAddressIndex: optionalNumber,
  }),
]);

export type EthBuildOptionsWithoutFeeOptions = t.TypeOf<
  typeof EthBuildOptionsWithoutFeeOptions
>;

// This is the same as Omit<EthBuildOptions, 'hopParams' | 'nonce'>
// but `omit` is not easily available in io-ts
export const EthFillNonceBuildOptions = t.partial({
  feeOptions: Optional(t.union([FeeOption, EIP1559FeeOptions])),
  receiveAddress: optionalString,
  senderAddressIndex: optionalNumber,
});

export type EthFillNonceBuildOptions = t.TypeOf<
  typeof EthFillNonceBuildOptions
>;

export const AllEthBuildOptions = t.union([
  EthBuildOptions,
  EthBuildOptionsWithoutFeeOptions,
  EthFillNonceBuildOptions,
]);

export type AllEthBuildOptions = t.TypeOf<typeof AllEthBuildOptions>;
