// @bitgo/public-types/src/schema/transactionRequest/intents/feeConfig.ts
import * as t from "io-ts";

/**
 * Fee dimensions for FLR P chain transactions
 */
enum FeeDimensions {
  Bandwidth = 0,
  DBRead = 1,
  DBWrite = 2,
  Compute = 3,
}

/**
 * Dimension value type
 */
type DimensionValue = number;

/**
 * Dimensions mapping for fee calculation
 */
type Dimensions = Record<FeeDimensions, DimensionValue>;

/**
 * Fee configuration for post-Etna FLR P chain transactions
 * @property {Dimensions} weights - Dimension weights for fee calculation
 * @property {bigint} maxCapacity - Maximum capacity value
 * @property {bigint} maxPerSecond - Maximum units per second
 * @property {bigint} targetPerSecond - Target units per second
 * @property {bigint} minPrice - Minimum gas price
 * @property {bigint} excessConversionConstant - Conversion constant for excess calculation
 */
export const FeeConfig = t.type({
  weights: t.record(t.string, t.number),
  maxCapacity: t.bigint,
  maxPerSecond: t.bigint,
  targetPerSecond: t.bigint,
  minPrice: t.bigint,
  excessConversionConstant: t.bigint,
});

export type FeeConfig = Omit<t.TypeOf<typeof FeeConfig>, "weights"> & {
  weights: Dimensions;
};
