// @bitgo/public-types/src/schema/transactionRequest/intents/flrpContext.ts
import * as t from "io-ts";
import { FeeConfig } from "./feeConfig";

/**
 * Context for FLR P chain transactions
 * @property {number} networkID - Network identifier
 * @property {string} hrp - Human-readable part for addresses
 * @property {string} xBlockchainID - X-Chain blockchain ID
 * @property {string} pBlockchainID - P-Chain blockchain ID
 * @property {string} cBlockchainID - C-Chain blockchain ID
 * @property {string} avaxAssetID - AVAX asset identifier
 * @property {bigint} baseTxFee - Base transaction fee
 * @property {bigint} createAssetTxFee - Create asset transaction fee
 * @property {bigint} createSubnetTxFee - Create subnet transaction fee
 * @property {bigint} transformSubnetTxFee - Transform subnet transaction fee
 * @property {bigint} createBlockchainTxFee - Create blockchain transaction fee
 * @property {bigint} addPrimaryNetworkValidatorFee - Add primary network validator fee
 * @property {bigint} addPrimaryNetworkDelegatorFee - Add primary network delegator fee
 * @property {bigint} addSubnetValidatorFee - Add subnet validator fee
 * @property {bigint} addSubnetDelegatorFee - Add subnet delegator fee
 * @property {FeeConfig} platformFeeConfig - Platform fee configuration (Post Etna)
 */
export const FlrpContext = t.type({
  networkID: t.number,
  hrp: t.string,
  xBlockchainID: t.string,
  pBlockchainID: t.string,
  cBlockchainID: t.string,
  avaxAssetID: t.string,
  baseTxFee: t.bigint,
  createAssetTxFee: t.bigint,
  createSubnetTxFee: t.bigint,
  transformSubnetTxFee: t.bigint,
  createBlockchainTxFee: t.bigint,
  addPrimaryNetworkValidatorFee: t.bigint,
  addPrimaryNetworkDelegatorFee: t.bigint,
  addSubnetValidatorFee: t.bigint,
  addSubnetDelegatorFee: t.bigint,
  platformFeeConfig: FeeConfig,
});

export type FlrpContext = t.TypeOf<typeof FlrpContext>;
