// @bitgo/public-types/src/schema/transactionRequest/intents/intentType.ts
import * as t from "io-ts";

export const intentTypes = {
  payment: t.literal("payment"),
  transferToken: t.literal("transferToken"),
  consolidate: t.literal("consolidate"),
  consolidateToken: t.literal("consolidateToken"),
  fanout: t.literal("fanout"),
  stake: t.literal("stake"),
  unstake: t.literal("unstake"),
  delegate: t.literal("delegate"),
  undelegate: t.literal("undelegate"),
  switchValidator: t.literal("switchValidator"),
  claim: t.literal("claim"),
  stakeClaimRewards: t.literal("stakeClaimRewards"),
  pledge: t.literal("pledge"),
  voteDelegation: t.literal("voteDelegation"),
  vote: t.literal("vote"),
  createAccount: t.literal("createAccount"),
  updateAccount: t.literal("updateAccount"),
  addTrustLine: t.literal("addTrustLine"),
  removeTrustLine: t.literal("removeTrustLine"),
  signMessage: t.literal("signMessage"),
  signTypedStructuredData: t.literal("signTypedStructuredData"),
  enableToken: t.literal("enableToken"),
  authorize: t.literal("authorize"),
  acceleration: t.literal("acceleration"),
  fillNonce: t.literal("fillNonce"),
  walletRecovery: t.literal("walletRecovery"),
  contractCall: t.literal("contractCall"),
  deactivate: t.literal("deactivate"),
  customTx: t.literal("customTx"),
  closeAssociatedTokenAccount: t.literal("closeAssociatedTokenAccount"),
  stakeWithCallData: t.literal("stakeWithCallData"),
  unstakeWithCallData: t.literal("unstakeWithCallData"),
  switchValidatorWithCallData: t.literal("switchValidatorWithCallData"),
  feeAddressTransfer: t.literal("feeAddressTransfer"),
  tokenApproval: t.literal("tokenApproval"),
  transferStake: t.literal("transferStake"),
  validatorRegistration: t.literal("validatorRegistration"),
  spotTransfer: t.literal("spotTransfer"),
  bridgeFunds: t.literal("bridgeFunds"),
  /**
   * Unstake and transfer funds from a staking wallet to a go account.
   */
  goUnstake: t.literal("goUnstake"),
  /**
   * Register a BTC delegation on the Babylon chain for Baby BTC staking.
   */
  createBtcDelegation: t.literal("createBtcDelegation"),
  transferAccept: t.literal("transferAccept"),
  transferReject: t.literal("transferReject"),
  transferAcknowledge: t.literal("transferAcknowledge"),
  transferOfferWithdrawn: t.literal("transferOfferWithdrawn"),
};

export const IntentType = t.keyof(intentTypes);

export type IntentType = t.TypeOf<typeof IntentType>;
