// @bitgo/public-types/src/schema/transactionRequest/intents/irysStakeIntent.ts
import * as t from "io-ts";
import { StakeIntent } from "./stakeIntent";

/**
 * Parameters for Irys commitment transactions.
 * These values originate from the Irys pricing APIs called by staking-service.
 */
export const IrysCommitmentParams = t.type({
  /** Commitment fee as a bigint string (wei), from Irys price API. */
  commitmentFee: t.string,
  /** Commitment value as a bigint string (wei), from Irys price API. */
  commitmentValue: t.string,
  /** 0x-prefixed hex Ethereum address of the signer. */
  signerAddress: t.string,
  /** Irys chain ID: 1270 (testnet) or 3282 (mainnet). */
  chainId: t.number,
});

export type IrysCommitmentParams = t.TypeOf<typeof IrysCommitmentParams>;

/**
 * @title Irys Stake Intent
 *
 * Intent for Irys STAKE commitment transactions.
 * STAKE locks tokens as collateral (done once per wallet).
 * Routes to BGM_IRYS (mechanism 86) via irys/signtx keyserver path.
 */
export const IrysStakeIntent = t.intersection([
  StakeIntent,
  t.type({
    commitmentParams: IrysCommitmentParams,
  }),
]);

export type IrysStakeIntent = t.TypeOf<typeof IrysStakeIntent>;
