// @bitgo/public-types/src/schema/transactionRequest/intents/solStakeIntent.ts
import * as t from "io-ts";
import { SolBuildOptions } from "./solBuildOptions";
import { StakeIntent } from "./stakeIntent";
import { BaseIntentWithAmount } from "./baseIntent";
import { Optional } from "../../../utils";
import { SolStakingType } from "./solStakingType";

/**
 * @title SOL Stake Intent
 */
export const SolStakeIntent = t.intersection([
  StakeIntent,
  SolBuildOptions,
  BaseIntentWithAmount,
  t.type({
    validatorAddress: t.string,
    /**
     * @deprecated Use `stakingType` instead.
     */
    isMarinade: Optional(t.boolean),
    /**
     * @todo Remove Optional once staking-service is updated.
     */
    stakingType: Optional(SolStakingType),
  }),
]);

export type SolStakeIntent = t.TypeOf<typeof SolStakeIntent>;

/**
 * @title SOL Stake Intent for Jito
 */
export const SolJitoStakeIntent = t.intersection([
  SolStakeIntent,
  t.type({
    stakingType: t.literal("JITO"),
    stakePoolConfig: t.type({
      poolMint: t.string,
      reserveStake: t.string,
      managerFeeAccount: t.string,
    }),
  }),
]);

export type SolJitoStakeIntent = t.TypeOf<typeof SolJitoStakeIntent>;
