// @bitgo/public-types/src/schema/transactionRequest/intents/solUnstakeIntent.ts
import * as t from "io-ts";
import { SolBuildOptions } from "./solBuildOptions";
import { Amount } from "./amount";
import { UnstakeIntent } from "./unstakeIntent";
import { Optional } from "../../../utils";
import { RecipientEntry } from "./recipientEntry";
import { SolStakingType } from "./solStakingType";

/**
 * @title SOL Unstake Intent
 */
export const SolUnstakeIntent = t.intersection([
  UnstakeIntent,
  SolBuildOptions,
  t.type({
    stakingAddress: t.string,
  }),
  t.partial({
    amount: Optional(Amount),
    remainingStakingAmount: Optional(Amount),
    /**
     * @deprecated Use `stakingType` instead.
     */
    isMarinade: Optional(t.boolean),
    /**
     * @todo Remove Optional once staking-service is updated.
     */
    stakingType: Optional(SolStakingType),
    recipients: Optional(t.array(RecipientEntry)),
  }),
]);

export type SolUnstakeIntent = t.TypeOf<typeof SolUnstakeIntent>;

/**
 * @title SOL Unstake Intent for Jito
 */
export const SolJitoUnstakeIntent = t.intersection([
  SolUnstakeIntent,
  t.type({
    stakingType: t.literal("JITO"),
    stakePoolConfig: t.type({
      poolMint: t.string,
      validatorListAccount: t.string,
      managerFeeAccount: t.string,
    }),
    validatorAddress: t.string,
  }),
]);

export type SolJitoUnstakeIntent = t.TypeOf<typeof SolJitoUnstakeIntent>;
