// @bitgo/public-types/src/schema/transactionRequest/intents/solVersionedCustomTxIntent.ts
import * as t from "io-ts";
import { SolBuildOptions } from "./solBuildOptions";
import { BaseIntent } from "./baseIntent";
import { intentTypes } from "./intentType";

/**
 * @title SOL Versioned Custom Transaction Intent
 */
export const SolVersionedCustomTxIntent = t.intersection([
  BaseIntent,
  SolBuildOptions,
  t.type({
    intentType: intentTypes.customTx,
    solVersionedTransactionData: t.intersection([
      t.type({
        versionedInstructions: t.array(
          t.type({
            programIdIndex: t.number,
            accountKeyIndexes: t.array(t.number),
            data: t.string,
          }),
        ),
        addressLookupTables: t.array(
          t.type({
            accountKey: t.string,
            writableIndexes: t.array(t.number),
            readonlyIndexes: t.array(t.number),
          }),
        ),
        staticAccountKeys: t.array(t.string),
        messageHeader: t.type({
          numRequiredSignatures: t.number,
          numReadonlySignedAccounts: t.number,
          numReadonlyUnsignedAccounts: t.number,
        }),
      }),
      t.partial({
        recentBlockhash: t.string,
      }),
    ]),
  }),
]);

export type SolVersionedCustomTxIntent = t.TypeOf<
  typeof SolVersionedCustomTxIntent
>;
