// @bitgo/public-types/src/schema/transactionRequest/intents/tonDelegateIntent.ts
import * as t from "io-ts";
import { BaseIntent, BaseIntentWithAmount } from "./baseIntent";
import { intentTypes } from "./intentType";
import { optionalString } from "../../../utils";
import { TonStakingType } from "./tonStakingType";

/**
 * @title TON Delegate Intent
 *
 * @description The intent for delegate/stake requests and transactions for TON
 *
 * @param {string} intentType The type of intent (always set to delegate)
 * @param {string} validatorAddress The address of the validator
 * @param {boolean} useSingleNominator Whether to use a single nominator pool, if false then a multi nominator pool will be used
 * @param {Amount} singleNominatorAmount The amount to delegate to the single nominator pool
 * @param {string} ownerAddress The owner wallet address of the single nominator pool, if applicable (optional)
 */
export const TonDelegateIntent = t.intersection([
  BaseIntent,
  BaseIntentWithAmount,
  t.type({
    intentType: intentTypes.delegate,
    validatorAddress: t.string,
    // TODO SC-4508: deprecate useSingleNominator in favor of tonStakingType
    useSingleNominator: t.boolean,
  }),
  t.partial({
    // TODO SC-4508: move this to the TonSingleNominatorDelegateIntent and deprecate it here
    ownerAddress: optionalString,
    tonStakingType: TonStakingType,
  }),
]);

export type TonDelegateIntent = t.TypeOf<typeof TonDelegateIntent>;
