// @bitgo/public-types/src/schema/transactionRequest/intents/tonUndelegateIntent.ts
import * as t from "io-ts";
import { BaseIntent } from "./baseIntent";
import { intentTypes } from "./intentType";
import { Amount } from "./amount";
import { TonStakingType } from "./tonStakingType";

/**
 * @title TON Undelegate Intent
 * @description The intent for undelegate/unstake requests and transactions for TON
 *
 * @param {string} intentType The type of intent (always set to undelegate)
 * @param {string} validatorAddress The address of the validator
 * @param {boolean} useSingleNominator Whether to use a single nominator pool, if false then a multi nominator pool will be used
 * @param {Amount} singleNominatorAmount The amount to delegate to the single nominator pool
 * @param {string} [tonStakingType] The type of TON staking (SINGLE_NOMINATOR, MULTI_NOMINATOR, TON_WHALES)
 */
export const TonUndelegateIntent = t.intersection([
  BaseIntent,
  t.type({
    intentType: intentTypes.undelegate,
    validatorAddress: t.string,
    useSingleNominator: t.boolean,
    singleNominatorAmount: Amount,
  }),
  t.partial({
    tonStakingType: TonStakingType,
  }),
]);

export type TonUndelegateIntent = t.TypeOf<typeof TonUndelegateIntent>;
