// @bitgo/public-types/src/schema/transactionRequest/intents/xdcStakingIntent.ts
import * as t from "io-ts";
import { BaseStakeIntent } from "./baseStakeIntent";
import { intentTypes } from "./intentType";
import { NonEmptyString } from "io-ts-types";
import { ValidatorRegistrationIntent } from "./validatorRegistrationIntent";
import { BaseIntentWithAmount } from "./baseIntent";

/**
 * @title XDC Upload KYC Intent
 *
 * Intent for uploading KYC hash to XDC Validator contract.
 *
 * @param {string} intentType - The type of intent (contractCall)
 * @param {string} stakingRequestId - The staking request identifier
 * @param {string} ipfsHash - The IPFS hash of the uploaded KYC document (e.g., "Qm...")
 * @param {string} validatorContractAddress - The address of the XDC Validator contract
 *
 */
export const XdcUploadKycIntent = t.intersection([
  BaseStakeIntent,
  t.type({
    intentType: intentTypes.contractCall,
    ipfsHash: NonEmptyString,
    validatorContractAddress: NonEmptyString,
  }),
]);

export type XdcUploadKycIntent = t.TypeOf<typeof XdcUploadKycIntent>;

/**
 * @title XDC Propose Intent
 *
 * Intent for proposing a validator candidate on XDC network.
 * This registers the validator and stakes XDC.
 * Must be done AFTER uploading KYC.
 *
 * @param {string} intentType - The type of intent (validatorRegistration)
 * @param {string} stakingRequestId - The staking request identifier
 * @param {Amount} amount - The amount of XDC to stake (must be >= minCandidateCap)
 * @param {string} candidateAddress - The address of the validator candidate
 * @param {string} validatorContractAddress - The address of the XDC Validator contract
 */
export const XdcProposeIntent = t.intersection([
  ValidatorRegistrationIntent,
  BaseIntentWithAmount,
  t.type({
    candidateAddress: NonEmptyString,
    validatorContractAddress: NonEmptyString,
  }),
]);

export type XdcProposeIntent = t.TypeOf<typeof XdcProposeIntent>;
