// @bitgo/public-types/src/schema/transactionRequest/intents/xdcUnstakingIntent.ts
import * as t from "io-ts";
import { BaseStakeIntent } from "./baseStakeIntent";
import { intentTypes } from "./intentType";
import { NonEmptyString } from "io-ts-types";
import { BaseIntentWithAmount } from "./baseIntent";

/**
 * @title XDC Resign Intent
 *
 * Intent for resigning a masternode candidate on XDC network.
 * This calls resign(address _candidate) on the system validator contract
 * to initiate the unstaking process.
 *
 * @param {string} intentType - The type of intent (undelegate)
 * @param {string} stakingRequestId - The staking request identifier
 * @param {string} candidateAddress - The address of the masternode candidate to resign
 * @param {string} validatorContractAddress - The address of the XDC Validator contract
 */
export const XdcResignIntent = t.intersection([
  BaseStakeIntent,
  t.type({
    intentType: intentTypes.undelegate,
    candidateAddress: NonEmptyString,
    validatorContractAddress: NonEmptyString,
  }),
]);

export type XdcResignIntent = t.TypeOf<typeof XdcResignIntent>;

/**
 * @title XDC Withdraw Intent
 *
 * Intent for withdrawing staked XDC after the lock period expires.
 * This calls withdraw(uint256 _blockNumber, uint256 _index) on the
 * system validator contract to reclaim the 10M XDC stake.
 *
 * @param {string} intentType - The type of intent (unstake)
 * @param {string} stakingRequestId - The staking request identifier
 * @param {string} blockNumber - The block number when the resignation was confirmed
 * @param {string} index - The withdrawal index
 * @param {string} validatorContractAddress - The address of the XDC Validator contract
 */
export const XdcWithdrawIntent = t.intersection([
  BaseStakeIntent,
  BaseIntentWithAmount,
  t.type({
    intentType: intentTypes.unstake,
    blockNumber: NonEmptyString,
    index: NonEmptyString,
    validatorContractAddress: NonEmptyString,
  }),
]);

export type XdcWithdrawIntent = t.TypeOf<typeof XdcWithdrawIntent>;
