// @bitgo/sdk-core/dist/src/account-lib/mpc/tss/ecdsa/rangeproof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyWithCheck = exports.verify = exports.proveWithCheck = exports.prove = exports.verifyNtildeProof = exports.generateNtildeProof = exports.generateNtilde = exports.randomPositiveCoPrimeTo = exports.generateSafePrimes = void 0;
/**
 * Zero Knowledge Range Proofs as described in (Two-party generation of DSA signatures)[1].
 * [1]: https://reitermk.github.io/papers/2004/IJIS.pdf
 */
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
Object.defineProperty(exports, "randomPositiveCoPrimeTo", { enumerable: true, get: function () { return sdk_lib_mpc_1.randomPositiveCoPrimeTo; } });
Object.defineProperty(exports, "generateSafePrimes", { enumerable: true, get: function () { return sdk_lib_mpc_1.generateSafePrimes; } });
const { generateNtilde, generateNtildeProof, verifyNtildeProof, prove, proveWithCheck, verify, verifyWithCheck } = sdk_lib_mpc_1.EcdsaRangeProof;
exports.generateNtilde = generateNtilde;
exports.generateNtildeProof = generateNtildeProof;
exports.verifyNtildeProof = verifyNtildeProof;
exports.prove = prove;
exports.proveWithCheck = proveWithCheck;
exports.verify = verify;
exports.verifyWithCheck = verifyWithCheck;
