// @bitgo/sdk-coin-ton/dist/src/lib/tokenTransferBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const tokenTransaction_1 = require("./tokenTransaction");
class TokenTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(coinConfig) {
        super(coinConfig);
        this._transaction = new tokenTransaction_1.TokenTransaction(coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.SendToken;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        this.transaction.transactionType = sdk_core_1.TransactionType.SendToken;
        this.transaction.fromRawTransaction(rawTransaction);
        return this.transaction;
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.transactionType = sdk_core_1.TransactionType.SendToken;
        await this.transaction.build();
        this.transaction.loadInputsAndOutputs();
        return this.transaction;
    }
    setForwardTonAmount(amount) {
        this.transaction.forwardTonAmount = amount;
        return this;
    }
    setSenderJettonWalletAddress(address) {
        this.transaction.senderJettonWalletAddress = address;
        return this;
    }
    setTonAmount(amount) {
        this.transaction.tonAmount = amount;
        return this;
    }
    // recipient method to handle both TON and token amounts
    recipient(address, senderJettonWalletAddress, tonAmount, jettonAmount, forwardTonAmount) {
        this._transaction.recipient = {
            address: address,
            amount: jettonAmount, // Jetton amount to be transferred
        };
        this._transaction.senderJettonWalletAddress = senderJettonWalletAddress; // The sender's Jetton wallet address
        this._transaction.tonAmount = tonAmount; // TON amount sent to the sender's Jetton wallet
        this._transaction.forwardTonAmount = forwardTonAmount; // TON amount to cover forward fees in case of notify transfer
        return this;
    }
    send(recipient) {
        this.transaction.recipient = recipient;
        return this;
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
