// @bitgo/sdk-coin-ton/dist/src/lib/tokenTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransaction = void 0;
const tonweb_1 = __importDefault(require("tonweb"));
const bn_js_1 = require("bn.js");
const sdk_core_1 = require("@bitgo/sdk-core");
const transaction_1 = require("./transaction");
const constants_1 = require("./constants");
class TokenTransaction extends transaction_1.Transaction {
    constructor(coinConfig) {
        super(coinConfig);
    }
    toJson() {
        const json = super.toJson();
        return {
            ...json,
            forwardTonAmount: this.forwardTonAmount,
            senderJettonWalletAddress: this.senderJettonWalletAddress,
            tonAmount: this.tonAmount,
        };
    }
    createJettonTransferPayload(jettonAmount, toAddress, forwardTonAmount, comment) {
        const forwardPayload = new tonweb_1.default.boc.Cell();
        if (comment) {
            forwardPayload.bits.writeUint(0, 32);
            forwardPayload.bits.writeString(comment);
        }
        const payload = new tonweb_1.default.boc.Cell();
        payload.bits.writeUint(constants_1.JETTON_TRANSFER_OPCODE, 32); // Store Jetton transfer op code
        payload.bits.writeUint(0, 64); // query_id
        payload.bits.writeCoins(new bn_js_1.BN(jettonAmount)); // Jetton amount to transfer
        payload.bits.writeAddress(new tonweb_1.default.Address(toAddress)); // recipient's TON wallet address
        payload.bits.writeAddress(new tonweb_1.default.Address(this.sender)); // response_destination for sending excess TON (excess is returned back to the sender)
        payload.bits.writeBit(false); // No custom payload
        payload.bits.writeCoins(new bn_js_1.BN(forwardTonAmount)); // forward_ton_amount to pay fees
        payload.bits.writeBit(true); // Forward payload exists as a reference
        payload.refs.push(forwardPayload); // Add forward payload as a reference
        return payload;
    }
    async build() {
        const jettonTransferPayload = this.createJettonTransferPayload(this.recipient.amount, this.recipient.address, this.forwardTonAmount, this.message);
        const signingMessage = this.createSigningMessage(constants_1.WALLET_ID, this.seqno, this.expireTime);
        const sendMode = 3;
        signingMessage.bits.writeUint8(sendMode);
        const outMsg = this.createOutMsg(this.senderJettonWalletAddress, this.tonAmount, jettonTransferPayload);
        signingMessage.refs.push(outMsg);
        this.unsignedMessage = Buffer.from(await signingMessage.hash()).toString('hex');
        const signature = this._signatures.length > 0 ? this._signatures[0] : Buffer.from(new Uint8Array(64)).toString('hex');
        const finalMessage = await this.createExternalMessage(signingMessage, this.seqno, signature);
        this.finalMessage = tonweb_1.default.utils.bytesToBase64(await finalMessage.toBoc(false));
        const originalTxId = tonweb_1.default.utils.bytesToBase64(await finalMessage.hash());
        this._id = originalTxId.replace(/\//g, '_').replace(/\+/g, '-');
    }
    fromRawTransaction(rawTransaction) {
        try {
            const cell = tonweb_1.default.boc.Cell.oneFromBoc(tonweb_1.default.utils.base64ToBytes(rawTransaction));
            const parsed = this.parseTransaction(cell);
            this.transactionType = sdk_core_1.TransactionType.SendToken;
            this.sender = parsed.fromAddress;
            this.recipient = { address: parsed.payload.jettonRecipient, amount: parsed.payload.jettonAmount };
            this.tonAmount = parsed.value;
            this.forwardTonAmount = parsed.payload.forwardTonAmount;
            this.senderJettonWalletAddress = parsed.toAddress;
            this.seqno = parsed.seqno;
            this.publicKey = parsed.publicKey;
            this.expireTime = parsed.expireAt;
            this.message = parsed.payload.message;
            this._signatures.push(parsed.signature);
            this.bounceable = parsed.bounce;
        }
        catch (e) {
            throw new Error('invalid raw transaction');
        }
    }
    /** @inheritDoc */
    explainTransaction() {
        const displayOrder = ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'withdrawAmount'];
        const outputs = [this.recipient];
        const outputAmount = this.recipient.amount;
        const withdrawAmount = this.withdrawAmount;
        return {
            displayOrder,
            id: this.id,
            outputs,
            outputAmount,
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: 'UNKNOWN' },
            withdrawAmount,
        };
    }
}
exports.TokenTransaction = TokenTransaction;
