// @bitgo/sdk-coin-ton/dist/src/lib/tonWhalesWithdrawalBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TonWhalesWithdrawalBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
const constants_1 = require("./constants");
class TonWhalesWithdrawalBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.TonWhalesWithdrawal;
    }
    /**
     * Sets the payload for the withdrawal request.
     * Structure: OpCode (32) + QueryId (64) + GasLimit (Coins) + UnstakeAmount (Coins)
     * * @param unstakeAmount The amount of NanoTON to unstake (inside payload)
     * @param unstakeAmount The amount to unstake
     * @param queryId Optional custom query ID
     */
    setWithdrawalMessage(unstakeAmount, queryId) {
        const qId = queryId || '0000000000000000';
        this.transaction.withdrawAmount = unstakeAmount;
        this.transaction.message = constants_1.TON_WHALES_WITHDRAW_OPCODE + qId + unstakeAmount;
        return this;
    }
    /**
     * Sets the message to withdraw EVERYTHING from the pool.
     * This sets the unstakeAmount to "0", which is the specific signal for full withdrawal.
     */
    setFullWithdrawalMessage(queryId) {
        return this.setWithdrawalMessage('0', queryId);
    }
    /**
     * Sets the value attached to the transaction (The Fees).
     * NOTE: This is NOT the unstake amount. This is the fee paid to the pool
     * to process the request (e.g. withdrawFee + receiptPrice).
     * * @param amount NanoTON amount to attach to the message
     */
    setForwardAmount(amount) {
        if (!this.transaction.recipient) {
            this.transaction.recipient = { address: '', amount: amount };
        }
        else {
            this.transaction.recipient.amount = amount;
        }
        return this;
    }
    send(recipient) {
        this.transaction.recipient = recipient;
        return this;
    }
    setMessage(msg) {
        throw new Error('Use setWithdrawalMessage for specific payload construction');
    }
}
exports.TonWhalesWithdrawalBuilder = TonWhalesWithdrawalBuilder;
