// @bitgo/sdk-coin-ton/dist/src/lib/tonWhalesVestingWithdrawBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TonWhalesVestingWithdrawBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
class TonWhalesVestingWithdrawBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.TonWhalesVestingWithdrawal;
    }
    setWithdrawMessage() {
        this.transaction.message = 'Withdraw';
        return this;
    }
    setForwardAmount(amount) {
        if (!this.transaction.recipient) {
            this.transaction.recipient = { address: '', amount: amount };
        }
        else {
            this.transaction.recipient.amount = amount;
        }
        return this;
    }
    send(recipient) {
        this.transaction.recipient = recipient;
        return this;
    }
    setMessage(msg) {
        throw new Error('Method not implemented. Use setWithdrawMessage() instead.');
    }
}
exports.TonWhalesVestingWithdrawBuilder = TonWhalesVestingWithdrawBuilder;
