// @bitgo/sdk-coin-ton/dist/src/lib/tonWhalesVestingDepositBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TonWhalesVestingDepositBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
class TonWhalesVestingDepositBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.TonWhalesVestingDeposit;
    }
    setDepositMessage() {
        this.transaction.message = 'Deposit';
        return this;
    }
    setDepositAmount(amount) {
        if (!this.transaction.recipient) {
            this.transaction.recipient = { address: '', amount: amount };
        }
        else {
            this.transaction.recipient.amount = amount;
        }
        return this;
    }
    send(recipient) {
        this.transaction.recipient = recipient;
        return this;
    }
    setMessage(msg) {
        throw new Error('Method not implemented. Use setDepositMessage() instead.');
    }
}
exports.TonWhalesVestingDepositBuilder = TonWhalesVestingDepositBuilder;
