// @bitgo/sdk-coin-ton/dist/src/lib/tonWhalesDepositBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TonWhalesDepositBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const transaction_1 = require("./transaction");
const constants_1 = require("./constants");
class TonWhalesDepositBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.TonWhalesDeposit;
    }
    setDepositMessage(queryId) {
        // Deposit payload is just OpCode + QueryId.
        // The Amount is in the transaction value (recipient.amount)
        // The Gas Limit is hardcoded in Transaction.ts build()
        const qId = queryId || '0000000000000000';
        this.transaction.message = constants_1.TON_WHALES_DEPOSIT_OPCODE + qId;
        return this;
    }
    setDepositAmount(amount) {
        if (!this.transaction.recipient) {
            this.transaction.recipient = { address: '', amount: amount };
        }
        else {
            this.transaction.recipient.amount = amount;
        }
        return this;
    }
    send(recipient) {
        this.transaction.recipient = recipient;
        return this;
    }
    setMessage(msg) {
        throw new Error('Method not implemented.');
    }
}
exports.TonWhalesDepositBuilder = TonWhalesDepositBuilder;
