// @bitgo/sdk-coin-ton/dist/src/lib/transactionBuilderFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transferBuilder_1 = require("./transferBuilder");
const singleNominatorWithdrawBuilder_1 = require("./singleNominatorWithdrawBuilder");
const transaction_1 = require("./transaction");
const tokenTransferBuilder_1 = require("./tokenTransferBuilder");
const tokenTransaction_1 = require("./tokenTransaction");
const tonWhalesDepositBuilder_1 = require("./tonWhalesDepositBuilder");
const tonWhalesWithdrawalBuilder_1 = require("./tonWhalesWithdrawalBuilder");
const tonWhalesVestingDepositBuilder_1 = require("./tonWhalesVestingDepositBuilder");
const tonWhalesVestingWithdrawBuilder_1 = require("./tonWhalesVestingWithdrawBuilder");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    from(raw) {
        let builder;
        try {
            const checkTx = new transaction_1.Transaction(this._coinConfig);
            checkTx.fromRawTransaction(raw);
            let tx;
            if (checkTx.type === sdk_core_1.TransactionType.SendToken) {
                // It's a token transaction, so use TokenTransaction for proper parsing
                tx = new tokenTransaction_1.TokenTransaction(this._coinConfig);
                tx.fromRawTransaction(raw);
            }
            else {
                // It's a regular transaction, use the already parsed one
                tx = checkTx;
            }
            switch (tx.type) {
                case sdk_core_1.TransactionType.Send:
                    builder = this.getTransferBuilder();
                    break;
                case sdk_core_1.TransactionType.SingleNominatorWithdraw:
                    builder = this.getSingleNominatorWithdrawBuilder();
                    break;
                case sdk_core_1.TransactionType.SendToken:
                    builder = this.getTokenTransferBuilder();
                    break;
                case sdk_core_1.TransactionType.TonWhalesDeposit:
                    builder = this.getTonWhalesDepositBuilder();
                    break;
                case sdk_core_1.TransactionType.TonWhalesWithdrawal:
                    builder = this.getTonWhalesWithdrawalBuilder();
                    break;
                case sdk_core_1.TransactionType.TonWhalesVestingDeposit:
                    builder = this.getTonWhalesVestingDepositBuilder();
                    break;
                case sdk_core_1.TransactionType.TonWhalesVestingWithdrawal:
                    builder = this.getTonWhalesVestingWithdrawBuilder();
                    break;
                default:
                    throw new sdk_core_1.InvalidTransactionError('unsupported transaction');
            }
            builder.from(raw);
            return builder;
        }
        catch (e) {
            throw e;
        }
    }
    /** @inheritdoc */
    getTransferBuilder() {
        return new transferBuilder_1.TransferBuilder(this._coinConfig);
    }
    /**
     * Returns a specific builder to create a TON withdraw transaction
     */
    getSingleNominatorWithdrawBuilder() {
        return new singleNominatorWithdrawBuilder_1.SingleNominatorWithdrawBuilder(this._coinConfig);
    }
    /**
     * Returns a specific builder to create a TON token transfer transaction
     */
    getTokenTransferBuilder() {
        return new tokenTransferBuilder_1.TokenTransferBuilder(this._coinConfig);
    }
    /** @inheritdoc */
    getWalletInitializationBuilder() {
        throw new Error('Method not implemented.');
    }
    getTonWhalesDepositBuilder() {
        return new tonWhalesDepositBuilder_1.TonWhalesDepositBuilder(this._coinConfig);
    }
    getTonWhalesWithdrawalBuilder() {
        return new tonWhalesWithdrawalBuilder_1.TonWhalesWithdrawalBuilder(this._coinConfig);
    }
    getTonWhalesVestingDepositBuilder() {
        return new tonWhalesVestingDepositBuilder_1.TonWhalesVestingDepositBuilder(this._coinConfig);
    }
    getTonWhalesVestingWithdrawBuilder() {
        return new tonWhalesVestingWithdrawBuilder_1.TonWhalesVestingWithdrawBuilder(this._coinConfig);
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
