// @bitgo/sdk-coin-ton/dist/src/lib/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const tonweb_1 = __importDefault(require("tonweb"));
const bn_js_1 = require("bn.js");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this.bounceable = false;
        this.fromAddressBounceable = true;
        this.toAddressBounceable = true;
        this.isV3ContractMessage = false;
    }
    canSign(key) {
        return false;
    }
    toBroadcastFormat() {
        return this.finalMessage;
    }
    toJson() {
        const otherFormat = new tonweb_1.default.Address(this.recipient.address).toString(true, true, !new tonweb_1.default.Address(this.recipient.address).isBounceable);
        return {
            id: this._id,
            sender: this.sender,
            destination: this.recipient.address,
            destinationAlias: otherFormat,
            amount: this.recipient.amount,
            withdrawAmount: this.withdrawAmount,
            seqno: this.seqno,
            sub_wallet_id: this.sub_wallet_id,
            expirationTime: this.expireTime,
            publicKey: this.publicKey,
            signature: this._signatures[0],
            bounceable: this.bounceable,
        };
    }
    get signablePayload() {
        return Buffer.from(this.unsignedMessage, 'hex');
    }
    /**
     * Set the transaction type.
     * @param {TransactionType} transactionType The transaction type to be set.
     */
    set transactionType(transactionType) {
        this._type = transactionType;
    }
    async build() {
        if (!this.sub_wallet_id) {
            if (this.isV3ContractMessage) {
                this.sub_wallet_id = constants_1.VESTING_CONTRACT_WALLET_ID;
            }
            else {
                this.sub_wallet_id = constants_1.WALLET_ID;
            }
        }
        const signingMessage = this.createSigningMessage(this.sub_wallet_id, this.seqno, this.expireTime);
        const sendMode = 3; // default sendMode
        signingMessage.bits.writeUint8(sendMode);
        const outMsg = this.createOutMsg(this.recipient.address, this.recipient.amount, this.message);
        signingMessage.refs.push(outMsg);
        this.unsignedMessage = Buffer.from(await signingMessage.hash()).toString('hex');
        const signature = this._signatures.length > 0 ? this._signatures[0] : Buffer.from(new Uint8Array(64)).toString('hex');
        const finalMessage = await this.createExternalMessage(signingMessage, this.seqno, signature);
        this.finalMessage = tonweb_1.default.utils.bytesToBase64(await finalMessage.toBoc(false));
        const originalTxId = tonweb_1.default.utils.bytesToBase64(await finalMessage.hash());
        this._id = originalTxId.replace(/\//g, '_').replace(/\+/g, '-');
    }
    createSigningMessage(walletId, seqno, expireAt) {
        const message = new tonweb_1.default.boc.Cell();
        message.bits.writeUint(walletId, 32);
        // expireAt should be set as per the provided arg value, regardless of the seqno
        message.bits.writeUint(expireAt, 32);
        message.bits.writeUint(seqno, 32);
        if (!this.isV3ContractMessage) {
            message.bits.writeUint(0, 8); // op
        }
        return message;
    }
    createOutMsg(address, amount, payload) {
        let payloadCell = new tonweb_1.default.boc.Cell();
        if (payload) {
            if (payload.refs) {
                // is Cell
                payloadCell = payload;
            }
            else if (typeof payload === 'string') {
                if (payload.length > 0) {
                    if (payload.length > 24 && payload.substring(0, 8) === constants_1.WITHDRAW_OPCODE) {
                        // payload is a withdraw txn message
                        const queryId = payload.substring(8, 24);
                        const withdrawAmount = payload.substring(24);
                        /* eslint-disable-next-line no-console */
                        console.log(`payload: ${payload}, queryId: ${queryId}, withdrawAmount: ${withdrawAmount}`);
                        payloadCell.bits.writeUint(parseInt(constants_1.WITHDRAW_OPCODE, 16), 32);
                        payloadCell.bits.writeUint(parseInt(queryId, 16), 64);
                        payloadCell.bits.writeCoins(new bn_js_1.BN(withdrawAmount));
                    }
                    else if (payload.length >= 26 && payload.substring(0, 10) === constants_1.TON_WHALES_DEPOSIT_OPCODE) {
                        const queryId = payload.substring(10, 26);
                        payloadCell.bits.writeUint(parseInt(constants_1.TON_WHALES_DEPOSIT_OPCODE, 10), 32);
                        payloadCell.bits.writeUint(parseInt(queryId, 16), 64);
                        // The Ton Whales protocol requires a specific 'gas limit' field in the payload
                        // structure (OpCode -> QueryId -> GasLimit -> Amount).
                        payloadCell.bits.writeCoins(tonweb_1.default.utils.toNano('1'));
                    }
                    else if (payload.length >= 26 && payload.substring(0, 10) === constants_1.TON_WHALES_WITHDRAW_OPCODE) {
                        const queryId = payload.substring(10, 26);
                        const amountStr = payload.substring(26);
                        payloadCell.bits.writeUint(parseInt(constants_1.TON_WHALES_WITHDRAW_OPCODE, 10), 32);
                        payloadCell.bits.writeUint(new bn_js_1.BN(queryId, 16), 64);
                        // The Ton Whales protocol requires a specific 'gas limit' field in the payload
                        // structure (OpCode -> QueryId -> GasLimit -> Amount).
                        // We hardcode 1 TON here to match the Deposit implementation and ensure
                        // sufficient gas for the pool to process the request.
                        payloadCell.bits.writeCoins(tonweb_1.default.utils.toNano('1'));
                        payloadCell.bits.writeCoins(new bn_js_1.BN(amountStr));
                    }
                    else {
                        payloadCell.bits.writeUint(0, 32);
                        payloadCell.bits.writeString(payload);
                    }
                }
            }
            else {
                payloadCell.bits.writeBytes(payload);
            }
        }
        const orderHeader = tonweb_1.default.Contract.createInternalMessageHeader(new tonweb_1.default.Address(address), new bn_js_1.BN(amount), true, this.bounceable);
        return tonweb_1.default.Contract.createCommonMsgInfo(orderHeader, undefined, payloadCell);
    }
    async createExternalMessage(signingMessage, seqno, signature) {
        const body = new tonweb_1.default.boc.Cell();
        body.bits.writeBytes(Buffer.from(signature, 'hex'));
        body.writeCell(signingMessage);
        let stateInit;
        if (seqno === 0 && !this.isV3ContractMessage) {
            const WalletClass = tonweb_1.default.Wallets.all['v4R2'];
            const wallet = new WalletClass(new tonweb_1.default.HttpProvider(), {
                publicKey: tonweb_1.default.utils.hexToBytes(this.publicKey),
                wc: 0,
            });
            const deploy = await wallet.createStateInit();
            stateInit = deploy.stateInit;
        }
        const header = tonweb_1.default.Contract.createExternalMessageHeader(this.sender);
        return tonweb_1.default.Contract.createCommonMsgInfo(header, stateInit, body);
    }
    loadInputsAndOutputs() {
        const outputs = [];
        const inputs = [];
        inputs.push({
            address: this.sender,
            value: this.recipient.amount,
            coin: this._coinConfig.name,
        });
        outputs.push({
            address: this.recipient.address,
            value: this.recipient.amount,
            coin: this._coinConfig.name,
        });
        this._outputs = outputs;
        this._inputs = inputs;
    }
    fromRawTransaction(rawTransaction) {
        try {
            const cell = tonweb_1.default.boc.Cell.oneFromBoc(tonweb_1.default.utils.base64ToBytes(rawTransaction));
            const parsed = this.parseTransaction(cell);
            this.sender = parsed.fromAddress;
            this.recipient = { address: parsed.toAddress, amount: parsed.value };
            this.withdrawAmount = parsed.withdrawAmount;
            this.seqno = parsed.seqno;
            this.publicKey = parsed.publicKey;
            this.expireTime = parsed.expireAt;
            this.message = parsed.payload;
            this._signatures.push(parsed.signature);
            this.bounceable = parsed.bounce;
            this.sub_wallet_id = parsed.walletId;
        }
        catch (e) {
            throw new Error('invalid raw transaction');
        }
    }
    /** @inheritDoc */
    explainTransaction() {
        const displayOrder = ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'withdrawAmount'];
        const outputs = [this.recipient];
        const outputAmount = this.recipient.amount;
        const withdrawAmount = this.withdrawAmount;
        return {
            displayOrder,
            id: this.id,
            outputs,
            outputAmount,
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: 'UNKNOWN' },
            withdrawAmount,
        };
    }
    parseTransaction(cell) {
        const slice = cell.beginParse();
        // header
        if (slice.loadUint(2).toNumber() !== 2)
            throw Error('invalid header');
        const externalSourceAddress = slice.loadAddress();
        if (externalSourceAddress !== null)
            throw Error('invalid externalSourceAddress');
        const externalDestAddress = slice.loadAddress();
        const externalImportFee = slice.loadCoins();
        if (!externalImportFee.eq(new bn_js_1.BN(0)))
            throw new Error('invalid externalImportFee');
        // stateInit
        let publicKey;
        if (slice.loadBit()) {
            if (slice.loadBit()) {
                const stateInit = slice.loadRef();
                stateInit.loadRef();
                const data = stateInit.loadRef();
                const seqno = data.loadUint(32).toNumber();
                if (seqno !== 0)
                    throw new Error('invalid seqno');
                const walletId = data.loadUint(32).toNumber();
                if (walletId !== constants_1.WALLET_ID)
                    throw new Error('invalid wallet id');
                const publicKeyBuf = new Uint8Array(32);
                for (let i = 0; i < publicKeyBuf.length; i++) {
                    publicKeyBuf[i] = data.loadUint(8);
                }
                publicKey = Buffer.from(publicKeyBuf).toString('hex');
            }
        }
        // body
        const bodySlice = slice.loadBit() ? slice.loadRef() : slice;
        return {
            fromAddress: externalDestAddress.toString(true, true, this.fromAddressBounceable),
            publicKey,
            ...this.parseTransactionBody(bodySlice),
        };
    }
    parseTransactionBody(slice) {
        const signature = Buffer.from(slice.loadBits(512)).toString('hex');
        // signing message
        const walletId = slice.loadUint(32).toNumber();
        if (walletId !== constants_1.WALLET_ID && walletId !== constants_1.VESTING_CONTRACT_WALLET_ID)
            throw new Error('invalid walletId');
        const expireAt = slice.loadUint(32).toNumber();
        const seqno = slice.loadUint(32).toNumber();
        const op = slice.loadUint(8).toNumber();
        if (op !== 3) {
            if (op !== 0)
                throw new Error('invalid op');
            const sendMode = slice.loadUint(8).toNumber();
            if (sendMode !== 3)
                throw new Error('invalid sendMode');
        }
        else {
            this.isV3ContractMessage = true;
        }
        let order = slice.loadRef();
        if (order.loadBit())
            throw Error('invalid internal header');
        if (!order.loadBit())
            throw Error('invalid ihrDisabled');
        const bounce = order.loadBit();
        if (order.loadBit())
            throw Error('invalid bounced');
        const sourceAddress = order.loadAddress();
        if (sourceAddress !== null)
            throw Error('invalid externalSourceAddress');
        const destAddress = order.loadAddress();
        const value = order.loadCoins().toString();
        if (order.loadBit())
            throw Error('invalid currencyCollection');
        const ihrFees = order.loadCoins();
        if (!ihrFees.eq(new bn_js_1.BN(0)))
            throw new Error('invalid ihrFees');
        const fwdFees = order.loadCoins();
        if (!fwdFees.eq(new bn_js_1.BN(0)))
            throw new Error('invalid fwdFees');
        const createdLt = order.loadUint(64);
        if (!createdLt.eq(new bn_js_1.BN(0)))
            throw new Error('invalid createdLt');
        const createdAt = order.loadUint(32);
        if (!createdAt.eq(new bn_js_1.BN(0)))
            throw new Error('invalid createdAt');
        // order stateInit
        if (order.loadBit()) {
            order.loadRef(); // don't parse stateInit
        }
        // order body
        let payload;
        let withdrawAmount;
        this.transactionType = sdk_core_1.TransactionType.Send;
        if (order.getFreeBits() > 0) {
            if (order.loadBit()) {
                order = order.loadRef();
            }
            if (order.getFreeBits() > 32) {
                const opcode = order.loadUint(32).toNumber();
                if (opcode === 0) {
                    const payloadBytes = order.loadBits(order.getFreeBits());
                    payload = new TextDecoder().decode(payloadBytes);
                    if (payload === 'Deposit') {
                        this.transactionType = sdk_core_1.TransactionType.TonWhalesVestingDeposit;
                    }
                    else if (payload === 'Withdraw') {
                        this.transactionType = sdk_core_1.TransactionType.TonWhalesVestingWithdrawal;
                    }
                }
                else if (opcode === 4096) {
                    const queryId = order.loadUint(64).toNumber();
                    withdrawAmount = order.loadCoins().toNumber().toString();
                    payload = constants_1.WITHDRAW_OPCODE + queryId.toString(16).padStart(16, '0') + withdrawAmount;
                    this.transactionType = sdk_core_1.TransactionType.SingleNominatorWithdraw;
                }
                else if (opcode === constants_1.JETTON_TRANSFER_OPCODE) {
                    this.transactionType = sdk_core_1.TransactionType.SendToken;
                    const queryId = order.loadUint(64).toNumber();
                    if (queryId !== 0)
                        throw new Error('invalid queryId for jetton transfer');
                    const jettonAmount = order.loadCoins();
                    if (!jettonAmount.gt(new bn_js_1.BN(0)))
                        throw new Error('invalid jettonAmount');
                    const jettonRecipient = order.loadAddress();
                    if (!jettonRecipient)
                        throw new Error('invalid jettonRecipient');
                    const forwarderAddress = order.loadAddress();
                    if (!forwarderAddress)
                        throw new Error('invalid forwarderAddress');
                    order.loadBit(); // skip bit
                    const forwardTonAmount = order.loadCoins();
                    if (!forwardTonAmount.gt(new bn_js_1.BN(0)))
                        throw new Error('invalid forwardTonAmount');
                    let message = '';
                    if (order.loadBit()) {
                        order = order.loadRef();
                        const messageOpcode = order.loadUint(32).toNumber();
                        if (messageOpcode !== 0)
                            throw new Error('invalid message opcode');
                        const messageBytes = order.loadBits(order.getFreeBits());
                        message = new TextDecoder().decode(messageBytes);
                    }
                    payload = {
                        jettonAmount: jettonAmount.toString(),
                        jettonRecipient: jettonRecipient.toString(true, true, this.toAddressBounceable),
                        forwarderAddress: forwarderAddress.toString(true, true, this.fromAddressBounceable),
                        forwardTonAmount: forwardTonAmount.toString(),
                        message: message,
                    };
                }
                else if (opcode === parseInt(constants_1.TON_WHALES_DEPOSIT_OPCODE, 10)) {
                    this.transactionType = sdk_core_1.TransactionType.TonWhalesDeposit;
                    const queryId = order.loadUint(64).toNumber();
                    // This is the gas limit, which must be read to advance the cursor
                    // We do not need to store it
                    order.loadCoins();
                    payload = constants_1.TON_WHALES_DEPOSIT_OPCODE + queryId.toString(16).padStart(16, '0');
                }
                else if (opcode === parseInt(constants_1.TON_WHALES_WITHDRAW_OPCODE, 10)) {
                    this.transactionType = sdk_core_1.TransactionType.TonWhalesWithdrawal;
                    const queryId = order.loadUint(64).toNumber();
                    order.loadCoins(); // Skip Gas (Hardcoded in builder)
                    const amount = order.loadCoins();
                    withdrawAmount = amount.toString(); // Decimal String
                    // Reconstruct Payload: Decimal Op + Hex Query + Decimal Amount
                    const queryHex = new bn_js_1.BN(queryId).toString(16).padStart(16, '0');
                    payload = constants_1.TON_WHALES_WITHDRAW_OPCODE + queryHex + withdrawAmount;
                }
                else {
                    payload = '';
                }
            }
        }
        return {
            toAddress: destAddress.toString(true, true, this.fromAddressBounceable),
            value,
            bounce,
            seqno,
            withdrawAmount,
            expireAt,
            payload,
            signature,
            walletId,
        };
    }
    parseTransferStateInit(slice) {
        if (slice === null)
            return {};
        slice.loadRef();
        const data = slice.loadRef();
        const seqno = data.loadUint(32).toNumber();
        if (seqno !== 0)
            throw new Error('invalid seqno');
        const walletId = data.loadUint(32).toNumber();
        if (walletId !== constants_1.WALLET_ID)
            throw new Error('invalid wallet id');
        const publicKey = new Uint8Array(32);
        for (let i = 0; i < publicKey.length; i++) {
            publicKey[i] = data.loadUint(8);
        }
        return {
            publicKey: Buffer.from(publicKey).toString('hex'),
        };
    }
}
exports.Transaction = Transaction;
