// @bitgo/sdk-coin-ton/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
exports.getFeeEstimate = getFeeEstimate;
const tonweb_1 = __importDefault(require("tonweb"));
const bn_js_1 = require("bn.js");
const sdk_core_1 = require("@bitgo/sdk-core");
const constants_1 = require("./constants");
class Utils {
    /** @inheritdoc */
    isValidAddress(address) {
        try {
            if (address.length != 48) {
                return false;
            }
            Buffer.from(address, 'base64');
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    isValidBlockId(hash) {
        try {
            return Buffer.from(hash, 'base64').length === 32;
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    isValidPrivateKey(key) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    isValidPublicKey(key) {
        return (0, sdk_core_1.isValidEd25519PublicKey)(key);
    }
    /** @inheritdoc */
    isValidSignature(signature) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    isValidTransactionId(txId) {
        try {
            return Buffer.from(txId, 'base64').length === 32;
        }
        catch (e) {
            return false;
        }
    }
    async getAddressFromPublicKey(publicKey, bounceable = true, isUserFriendly = true) {
        const tonweb = new tonweb_1.default(new tonweb_1.default.HttpProvider(''));
        const WalletClass = tonweb.wallet.all['v4R2'];
        const wallet = new WalletClass(tonweb.provider, {
            publicKey: tonweb_1.default.utils.hexToBytes(publicKey),
            wc: 0,
        });
        const address = await wallet.getAddress();
        return address.toString(isUserFriendly, true, bounceable);
    }
    getAddress(address, bounceable = true) {
        if (bounceable) {
            return new tonweb_1.default.Address(address).isBounceable
                ? address
                : new tonweb_1.default.Address(address).toString(true, true, bounceable);
        }
        else {
            return new tonweb_1.default.Address(address).isBounceable
                ? new tonweb_1.default.Address(address).toString(true, true, bounceable)
                : address;
        }
    }
    async getMessageHashFromData(data) {
        const cell = tonweb_1.default.boc.Cell.oneFromBoc(tonweb_1.default.utils.base64ToBytes(data));
        // this is need to be confirmed by ton team
        const message = cell.refs[0].refs[0];
        const hash = tonweb_1.default.utils.bytesToBase64(await message.hash());
        return hash.toString();
    }
    getRawWalletAddressFromCell(data) {
        const cell = tonweb_1.default.boc.Cell.oneFromBoc(tonweb_1.default.utils.base64ToBytes(data));
        const slice = cell.beginParse();
        const address = slice.loadAddress();
        return address.toString();
    }
    async getVestingContractAddress(params) {
        const cell = new tonweb_1.default.boc.Cell();
        cell.bits.writeUint(0, 32); //seq no
        cell.bits.writeUint(params.subWalletId, 32);
        cell.bits.writeBytes(tonweb_1.default.utils.hexToBytes(params.publicKeyHex));
        cell.bits.writeUint(0, 1); // empty whitelist
        const vestingParamsCell = new tonweb_1.default.boc.Cell();
        vestingParamsCell.bits.writeUint(params.vestingStartTime, 64);
        vestingParamsCell.bits.writeUint(params.vestingTotalDuration, 32);
        vestingParamsCell.bits.writeUint(params.unlockPeriod, 32);
        vestingParamsCell.bits.writeUint(params.cliffDuration, 32);
        vestingParamsCell.bits.writeCoins(new bn_js_1.BN(params.vestingTotalAmount.toString()));
        const senderAddress = new tonweb_1.default.Address(params.vestingSenderAddress);
        vestingParamsCell.bits.writeAddress(senderAddress);
        const ownerAddress = new tonweb_1.default.Address(params.ownerAddress);
        vestingParamsCell.bits.writeAddress(ownerAddress);
        cell.refs.push(vestingParamsCell);
        const contractCodeCell = tonweb_1.default.boc.Cell.oneFromBoc(tonweb_1.default.utils.base64ToBytes(constants_1.VESTING_CONTRACT_CODE_B64));
        const stateInit = tonweb_1.default.Contract.createStateInit(contractCodeCell, cell);
        const stateInitHash = await stateInit.hash();
        const contractAddress = new tonweb_1.default.Address('0:' + tonweb_1.default.utils.bytesToHex(stateInitHash));
        return contractAddress.toString(true, true, true);
    }
    /**
     * Converts a TON address into a Base64 encoded BOC (Bag of Cells) string.
     * This is required when passing an address as a 'slice' argument to `runGetMethod`.
     * @param address - The TON address (friendly or raw)
     * @returns Base64 string of the serialized address cell
     */
    async getAddressBoc(address) {
        const cell = new tonweb_1.default.boc.Cell();
        cell.bits.writeAddress(new tonweb_1.default.Address(address));
        const boc = await cell.toBoc(false);
        return tonweb_1.default.utils.bytesToBase64(boc);
    }
}
exports.Utils = Utils;
const DUMMY_PRIVATE_KEY = '43e8594854cb53947c4a1a2fab926af11e123f6251dcd5bd0dfb100604186430'; // This dummy private key is used only for fee estimation
/**
 * Function to estimate the fee for a transaction.
 * This function uses the dummy private key exclusively for fee estimation.
 * @param wallet - The wallet instance.
 * @param toAddress - The destination address.
 * @param amount - The amount to transfer.
 * @param seqno - The sequence number for the transaction.
 * @returns The estimated fee for the transaction.
 */
async function getFeeEstimate(wallet, toAddress, amount, seqno) {
    try {
        const secretKey = tonweb_1.default.utils.stringToBytes(DUMMY_PRIVATE_KEY);
        const feeEstimate = await wallet.methods
            .transfer({
            secretKey,
            toAddress,
            amount,
            seqno,
            sendMode: 1,
        })
            .estimateFee();
        return feeEstimate;
    }
    catch (error) {
        throw new Error(`Failed to estimate fee: ${error.message}`);
    }
}
const utils = new Utils();
exports.default = utils;
