// @bitgo/sdk-coin-apt/dist/src/lib/utils/validation.js
"use strict";
/**
 * Validation utilities for Aptos transactions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateModuleName = validateModuleName;
exports.validateFunctionName = validateFunctionName;
exports.isValidModuleName = isValidModuleName;
exports.isValidFunctionName = isValidFunctionName;
/**
 * Core validation logic for module names
 *
 * @param {string} moduleName - Module name to validate
 * @returns {ValidationResult} Validation result with isValid flag and optional error message
 */
function validateModuleNameCore(moduleName) {
    if (!moduleName || typeof moduleName !== 'string') {
        return {
            isValid: false,
            errorMessage: 'Module name is required and must be a non-empty string',
        };
    }
    // Aptos module name format: address::module_name
    // Supports both SHORT (0x1) and LONG (0x0000...0001) address formats
    const moduleNamePattern = /^0x[a-fA-F0-9]{1,64}::[a-zA-Z_][a-zA-Z0-9_]*$/;
    if (!moduleNamePattern.test(moduleName)) {
        return {
            isValid: false,
            errorMessage: `Invalid module name format: "${moduleName}". Expected format: "0xaddress::module_name" (hex addresses only)`,
        };
    }
    return { isValid: true };
}
/**
 * Core validation logic for function names
 *
 * @param {string} functionName - Function name to validate
 * @returns {ValidationResult} Validation result with isValid flag and optional error message
 */
function validateFunctionNameCore(functionName) {
    if (!functionName || typeof functionName !== 'string') {
        return {
            isValid: false,
            errorMessage: 'Function name is required and must be a non-empty string',
        };
    }
    // Aptos function name pattern: valid identifier (letters, numbers, underscores, starting with letter/underscore)
    const functionNamePattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/;
    if (!functionNamePattern.test(functionName)) {
        return {
            isValid: false,
            errorMessage: `Invalid function name format: "${functionName}". Function names must be valid identifiers (letters, numbers, underscores, starting with letter or underscore)`,
        };
    }
    return { isValid: true };
}
/**
 * Validate module name format (throwing version)
 *
 * @param {string} moduleName - Module name to validate
 * @throws {Error} If module name format is invalid
 */
function validateModuleName(moduleName) {
    const result = validateModuleNameCore(moduleName);
    if (!result.isValid) {
        throw new Error(result.errorMessage);
    }
}
/**
 * Validate function name format (throwing version)
 *
 * @param {string} functionName - Function name to validate
 * @throws {Error} If function name format is invalid
 */
function validateFunctionName(functionName) {
    const result = validateFunctionNameCore(functionName);
    if (!result.isValid) {
        throw new Error(result.errorMessage);
    }
}
/**
 * Check if a module name matches the expected pattern (non-throwing version)
 *
 * @param {string} moduleName - Module name to check
 * @returns {boolean} True if valid, false otherwise
 */
function isValidModuleName(moduleName) {
    return validateModuleNameCore(moduleName).isValid;
}
/**
 * Check if a function name matches the expected pattern (non-throwing version)
 *
 * @param {string} functionName - Function name to check
 * @returns {boolean} True if valid, false otherwise
 */
function isValidFunctionName(functionName) {
    return validateFunctionNameCore(functionName).isValid;
}
