// @bitgo/sdk-coin-icp/dist/test/unit/transactionBuilder/transactionBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_core_1 = require("@bitgo/sdk-core");
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const sinon_1 = __importDefault(require("sinon"));
const iface_1 = require("../../../src/lib/iface");
const testData = __importStar(require("../../resources/icp"));
const getBuilderFactory_1 = require("../getBuilderFactory");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
describe('ICP Transaction Builder', async () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('ticp');
    let txBuilder;
    let txn;
    beforeEach(async function () {
        sinon_1.default.stub(utils_1.default, 'getMetaData').returns({
            metaData: testData.MetaDataWithMemo,
            ingressEndTime: testData.MetaDataWithMemo.ingress_end ?? 0,
        });
        sinon_1.default.stub(utils_1.default, 'validateExpireTime').returns(true);
        txBuilder = factory.getTransferBuilder();
        txBuilder.sender(testData.Accounts.account1.address, testData.Accounts.account1.publicKey);
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.amount('10');
        txBuilder.memo(testData.MetaDataWithMemo.memo);
        await txBuilder.build();
        txn = txBuilder.transaction;
    });
    afterEach(function () {
        sinon_1.default.restore();
    });
    it('should parse an unsigned transaction', async () => {
        const parsedUnsignedTxn = await factory.parseTransaction(testData.PayloadsData.unsigned_transaction, false);
        txn.should.be.an.Object();
        should_1.default.deepEqual(parsedUnsignedTxn, testData.ParsedUnsignedTransaction);
    });
    it('should parse an signed transaction', async () => {
        const parsedSignedTxn = await factory.parseTransaction(testData.SignedTransaction, true);
        txn.should.be.an.Object();
        should_1.default.deepEqual(parsedSignedTxn, testData.ParsedSignedTransaction);
    });
    it('start and build a transfer tx', async () => {
        const icpTransaction = txBuilder.transaction.icpTransaction;
        const payloadsData = txBuilder.transaction.payloadsData;
        should_1.default.equal(icpTransaction.metadata.memo, testData.MetaDataWithMemo.memo);
        should_1.default.equal(icpTransaction.operations[0].account.address, testData.Accounts.account1.address);
        should_1.default.equal(icpTransaction.operations[1].account.address, testData.Accounts.account2.address);
        should_1.default.equal(icpTransaction.operations[0].amount.value, '-10');
        should_1.default.equal(icpTransaction.operations[1].amount.value, '10');
        should_1.default.equal(icpTransaction.operations[2].amount.value, '-10000');
        should_1.default.equal(icpTransaction.public_keys[0].hex_bytes, testData.Accounts.account1.publicKey);
        payloadsData.payloads.should.be.an.Array();
        payloadsData.payloads.length.should.equal(1);
        should_1.default.deepEqual(txBuilder.payloadData(), testData.PayloadsData);
        should_1.default.deepEqual(payloadsData.payloads, testData.PayloadsData.payloads);
    });
    it('should fail to build a txn without sender', async () => {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.amount('10');
        txBuilder.memo(123456);
        await txBuilder.build().should.rejectedWith('sender is required before building');
    });
    it('should fail to build a txn with incorrect memo', async () => {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sender(testData.Accounts.account1.address, testData.Accounts.account1.publicKey);
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.amount('10');
        const incorrectMemos = ['abcd', undefined];
        for (const memo of incorrectMemos) {
            assert_1.default.throws(() => txBuilder.memo(Number(memo)), sdk_core_1.BuildTransactionError, `Invalid memo: ${Number(memo)}`);
        }
    });
    it('should fail to build a txn without amount', async () => {
        const txBuilder = factory.getTransferBuilder();
        txBuilder.sender(testData.Accounts.account1.address, testData.Accounts.account1.publicKey);
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.memo(123456);
        await txBuilder.build().should.rejectedWith('amount is required before building');
    });
    it('should build a signed txn and give txn in broadcast format', async () => {
        txn.addSignature(testData.Signatures);
        txBuilder.combine();
        const signedTxn = txBuilder.transaction.signedTransaction;
        signedTxn.should.be.a.String();
        should_1.default.equal(signedTxn, testData.SignedTransaction);
        const broadcastTxn = txBuilder.transaction.toBroadcastFormat();
        broadcastTxn.should.be.a.String();
        should_1.default.equal(broadcastTxn, signedTxn);
    });
    it('should sign a txn and then give txn in broadcast format', async () => {
        const baseKey = { key: testData.Accounts.account1.secretKey };
        txBuilder.sign(baseKey);
        should_1.default.deepEqual(txBuilder.signaturePayload(), testData.Signatures);
        txBuilder.combine();
        const signedTxn = txBuilder.transaction.signedTransaction;
        signedTxn.should.be.a.String();
        should_1.default.equal(signedTxn, testData.SignedTransaction);
        const transactionHash = txBuilder.transaction.id;
        should_1.default.equal(transactionHash, testData.OnChainTransactionHash);
        const broadcastTxn = txBuilder.transaction.toBroadcastFormat();
        broadcastTxn.should.be.a.String();
        should_1.default.equal(broadcastTxn, signedTxn);
    });
    it('should generate a correct txn hash', async () => {
        const unsignedTxn = txBuilder.transaction.unsignedTransaction;
        unsignedTxn.should.be.a.String();
        const payloadsData = txBuilder.transaction.payloadsData;
        const serializedTxFormat = {
            serializedTxHex: payloadsData,
            publicKey: testData.Accounts.account1.publicKey,
        };
        const serializedTxHex = Buffer.from(JSON.stringify(serializedTxFormat), 'utf-8').toString('hex');
        await txn.fromRawTransaction(serializedTxHex);
        const transactionHash = txBuilder.transaction.id;
        should_1.default.equal(transactionHash, testData.OnChainTransactionHash);
    });
    it('should build a txn then parse it and then again build', async () => {
        sinon_1.default.restore(); // do not stub getMetaData
        txBuilder = factory.getTransferBuilder();
        txBuilder.sender(testData.Accounts.account1.address, testData.Accounts.account1.publicKey);
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.amount('10');
        txBuilder.memo(testData.MetaDataWithMemo.memo);
        txBuilder.ingressEnd(1904384564000000000n);
        await txBuilder.build();
        txn = txBuilder.transaction;
        const unsignedTxn = txBuilder.transaction.unsignedTransaction;
        unsignedTxn.should.be.a.String();
        const payloadsData = txBuilder.transaction.payloadsData;
        const serializedTxFormat = {
            serializedTxHex: payloadsData,
            publicKey: testData.Accounts.account1.publicKey,
        };
        const serializedTxHex = Buffer.from(JSON.stringify(serializedTxFormat), 'utf-8').toString('hex');
        await txn.fromRawTransaction(serializedTxHex);
        const baseKey = { key: testData.Accounts.account1.secretKey };
        txBuilder.sign(baseKey);
        txBuilder.combine();
        const signedTxn = txBuilder.transaction.signedTransaction;
        signedTxn.should.be.a.String();
        txBuilder.transaction.icpTransaction.metadata.ingress_start.should.equal(Number(1904384564000000000n) - iface_1.MAX_INGRESS_TTL);
    });
});
describe('ICP Transaction Builder with default memo as 0', async () => {
    const factory = (0, getBuilderFactory_1.getBuilderFactory)('ticp');
    let txBuilder;
    let txn;
    beforeEach(async function () {
        sinon_1.default.stub(utils_1.default, 'getMetaData').returns({
            metaData: testData.MetaDataWithDefaultMemo,
            ingressEndTime: testData.MetaDataWithDefaultMemo.ingress_end ?? 0,
        });
        txBuilder = factory.getTransferBuilder();
        testData.ParsedUnsignedTransaction.metadata.memo = iface_1.DEFAULT_MEMO;
        testData.ParsedSignedTransaction.metadata.memo = iface_1.DEFAULT_MEMO;
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.sender(testData.Accounts.account1.address, testData.Accounts.account1.publicKey);
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.amount('10');
        await txBuilder.build();
        txn = txBuilder.transaction;
    });
    afterEach(function () {
        sinon_1.default.restore();
    });
    it('should parse an unsigned transaction with default memo', async () => {
        const parsedUnsignedTxn = await factory.parseTransaction(testData.PayloadsDataWithDefaultMemo.unsigned_transaction, false);
        txn.should.be.an.Object();
        should_1.default.deepEqual(parsedUnsignedTxn, testData.ParsedUnsignedTransaction);
    });
    it('should parse an signed transaction with default memo', async () => {
        const parsedSignedTxn = await factory.parseTransaction(testData.SignedTransactionWithDefaultMemo, true);
        txn.should.be.an.Object();
        should_1.default.deepEqual(parsedSignedTxn, testData.ParsedSignedTransaction);
    });
    it('start and build a transfer tx with default memo', async () => {
        const icpTransaction = txBuilder.transaction.icpTransaction;
        const payloadsData = txBuilder.transaction.payloadsData;
        should_1.default.equal(icpTransaction.metadata.memo, iface_1.DEFAULT_MEMO);
        should_1.default.equal(icpTransaction.operations[0].account.address, testData.Accounts.account1.address);
        should_1.default.equal(icpTransaction.operations[1].account.address, testData.Accounts.account2.address);
        should_1.default.equal(icpTransaction.operations[0].amount.value, '-10');
        should_1.default.equal(icpTransaction.operations[1].amount.value, '10');
        should_1.default.equal(icpTransaction.operations[2].amount.value, '-10000');
        should_1.default.equal(icpTransaction.public_keys[0].hex_bytes, testData.Accounts.account1.publicKey);
        payloadsData.payloads.should.be.an.Array();
        payloadsData.payloads.length.should.equal(1);
        should_1.default.deepEqual(txBuilder.payloadData(), testData.PayloadsDataWithDefaultMemo);
        should_1.default.deepEqual(payloadsData.payloads, testData.PayloadsDataWithDefaultMemo.payloads);
    });
    it('should build a signed txn with default memo and give txn in broadcast format', async () => {
        txn.addSignature(testData.SignaturesWithDefaultMemo);
        txBuilder.combine();
        const signedTxn = txBuilder.transaction.signedTransaction;
        signedTxn.should.be.a.String();
        should_1.default.equal(signedTxn, testData.SignedTransactionWithDefaultMemo);
        const broadcastTxn = txBuilder.transaction.toBroadcastFormat();
        broadcastTxn.should.be.a.String();
        should_1.default.equal(broadcastTxn, signedTxn);
    });
    it('should sign a txn with default memo and then give txn in broadcast format', async () => {
        const baseKey = { key: testData.Accounts.account1.secretKey };
        txBuilder.sign(baseKey);
        should_1.default.deepEqual(txBuilder.signaturePayload(), testData.SignaturesWithDefaultMemo);
        txBuilder.combine();
        const signedTxn = txBuilder.transaction.signedTransaction;
        signedTxn.should.be.a.String();
        should_1.default.equal(signedTxn, testData.SignedTransactionWithDefaultMemo);
        const broadcastTxn = txBuilder.transaction.toBroadcastFormat();
        broadcastTxn.should.be.a.String();
        should_1.default.equal(broadcastTxn, signedTxn);
    });
    it('should build a txn without memo then parse it and then again build', async () => {
        sinon_1.default.restore(); // do not stub getMetaData
        txBuilder = factory.getTransferBuilder();
        txBuilder.sender(testData.Accounts.account1.address, testData.Accounts.account1.publicKey);
        txBuilder.receiverId(testData.Accounts.account2.address);
        txBuilder.amount('10');
        await txBuilder.build();
        txn = txBuilder.transaction;
        const unsignedTxn = txBuilder.transaction.unsignedTransaction;
        unsignedTxn.should.be.a.String();
        const payloadsData = txBuilder.transaction.payloadsData;
        const serializedTxFormat = {
            serializedTxHex: payloadsData,
            publicKey: testData.Accounts.account1.publicKey,
        };
        const serializedTxHex = Buffer.from(JSON.stringify(serializedTxFormat), 'utf-8').toString('hex');
        await txn.fromRawTransaction(serializedTxHex);
        const baseKey = { key: testData.Accounts.account1.secretKey };
        txBuilder.sign(baseKey);
        txBuilder.combine();
        const signedTxn = txBuilder.transaction.signedTransaction;
        signedTxn.should.be.a.String();
    });
});
