// @bitgo/sdk-coin-icp/dist/test/unit/transactionBuilder/transactionRecover.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const testData = __importStar(require("../../resources/icp"));
const sinon_1 = __importDefault(require("sinon"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const nock_1 = __importDefault(require("nock"));
const index_1 = require("../../../src/index");
const icpAgent_1 = require("../../../src/lib/icpAgent");
const iface_1 = require("../../../src/lib/iface");
const principal_1 = require("@dfinity/principal");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const utils_1 = __importDefault(require("../../../src/lib/utils"));
describe('ICP transaction recovery', async () => {
    let bitgo;
    let recoveryParams;
    let icp;
    let broadcastEndpoint;
    let broadcastResponse;
    let nodeUrl;
    // Helper functions for setting up stubs
    const setupDefaultStubs = () => {
        sinon_1.default.stub(icpAgent_1.IcpAgent.prototype, 'getBalance').resolves((0, bignumber_js_1.default)(1000000000));
        sinon_1.default.stub(icpAgent_1.IcpAgent.prototype, 'getFee').resolves((0, bignumber_js_1.default)(10000));
        sinon_1.default.stub(utils_1.default, 'getMetaData').returns({
            metaData: testData.MetaDataWithDefaultMemo,
            ingressEndTime: testData.MetaDataWithDefaultMemo.ingress_end ?? 0,
        });
        sinon_1.default.stub(icp, 'signatures').returns(testData.RecoverTransactionSignatureWithDefaultMemo);
    };
    const setupMemoStubs = () => {
        sinon_1.default.stub(icpAgent_1.IcpAgent.prototype, 'getBalance').resolves((0, bignumber_js_1.default)(1000000000));
        sinon_1.default.stub(icpAgent_1.IcpAgent.prototype, 'getFee').resolves((0, bignumber_js_1.default)(10000));
        sinon_1.default.stub(utils_1.default, 'getMetaData').returns({
            metaData: testData.MetaDataWithMemo,
            ingressEndTime: testData.MetaDataWithMemo.ingress_end ?? 0,
        });
        sinon_1.default.stub(icp, 'signatures').returns(testData.RecoverTransactionSignatureWithMemo);
    };
    const setupLowBalanceStubs = () => {
        sinon_1.default.stub(icpAgent_1.IcpAgent.prototype, 'getBalance').resolves((0, bignumber_js_1.default)(10));
        sinon_1.default.stub(icpAgent_1.IcpAgent.prototype, 'getFee').resolves((0, bignumber_js_1.default)(10000));
        sinon_1.default.stub(utils_1.default, 'getMetaData').returns({
            metaData: testData.MetaDataWithDefaultMemo,
            ingressEndTime: testData.MetaDataWithDefaultMemo.ingress_end ?? 0,
        });
    };
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
        bitgo.safeRegister('icp', index_1.Icp.createInstance);
        bitgo.initializeTestVars();
        recoveryParams = {
            userKey: testData.WRWRecovery.userKey,
            backupKey: testData.WRWRecovery.backupKey,
            walletPassphrase: testData.WRWRecovery.walletPassphrase,
            recoveryDestination: testData.Accounts.account2.address,
        };
        icp = bitgo.coin('icp');
        nodeUrl = icp.getPublicNodeUrl();
        const principal = principal_1.Principal.fromUint8Array(iface_1.LEDGER_CANISTER_ID);
        const canisterIdHex = principal.toText();
        broadcastEndpoint = `/api/v3/canister/${canisterIdHex}/call`;
        broadcastResponse = Buffer.from(testData.PublicNodeApiBroadcastResponse, 'hex');
    });
    beforeEach(function () {
        setupDefaultStubs();
        // Set up default successful nock response
        (0, nock_1.default)(nodeUrl).post(broadcastEndpoint).reply(200, broadcastResponse);
    });
    afterEach(function () {
        recoveryParams = {
            userKey: testData.WRWRecovery.userKey,
            backupKey: testData.WRWRecovery.backupKey,
            walletPassphrase: testData.WRWRecovery.walletPassphrase,
            recoveryDestination: testData.Accounts.account2.address,
        };
        nock_1.default.cleanAll();
        sinon_1.default.restore();
    });
    it('should recover a transaction with default memo successfully', async () => {
        const recoverTxn = await icp.recover(recoveryParams);
        recoverTxn.id.should.be.a.String();
        should_1.default.equal(recoverTxn.id, testData.TxnIdWithDefaultMemo);
    });
    it('should recover a transaction with memo successfully', async () => {
        sinon_1.default.restore();
        setupMemoStubs();
        recoveryParams.memo = testData.MetaDataWithMemo.memo;
        const recoverTxn = await icp.recover(recoveryParams);
        recoverTxn.id.should.be.a.String();
        should_1.default.equal(recoverTxn.id, testData.TxnIdWithMemo);
    });
    it('should recover a unsigned sweep transaction successfully', async () => {
        sinon_1.default.restore();
        setupMemoStubs();
        const unsignedSweepRecoveryParams = {
            bitgoKey: '0310768736a005ea5364e1b5b5288cf553224dd28b2df8ced63b72a8020478967f05ec5bce1f26cd7eb009a4bea445bb55c2f54a30f2706c1a3747e8df2d288829',
            recoveryDestination: testData.Accounts.account2.address,
        };
        const recoverTxn = await icp.recover(unsignedSweepRecoveryParams);
        recoverTxn.txHex.should.be.a.String();
        should_1.default.equal(recoverTxn.txHex, testData.UnsignedSweepTransaction);
    });
    it('should failed to recover a unsigned sweep transaction with wrong bitgo key', async () => {
        sinon_1.default.restore();
        setupMemoStubs();
        const unsignedSweepRecoveryParams = {
            bitgoKey: 'testKey',
            recoveryDestination: testData.Accounts.account2.address,
        };
        await icp
            .recover(unsignedSweepRecoveryParams)
            .should.rejectedWith('Error during ICP recovery: Cannot convert 0x to a BigInt');
    });
    it('should failed to recover recover a unsigned sweep transaction without bitgo key', async () => {
        sinon_1.default.restore();
        setupMemoStubs();
        const unsignedSweepRecoveryParams = {
            recoveryDestination: testData.Accounts.account2.address,
        };
        await icp.recover(unsignedSweepRecoveryParams).should.rejectedWith('Error during ICP recovery: missing bitgoKey');
    });
    it('should fail to recover if broadcast API fails', async () => {
        nock_1.default.cleanAll();
        (0, nock_1.default)(nodeUrl).post(broadcastEndpoint).reply(500, 'Internal Server Error');
        recoveryParams.memo = 0;
        await icp
            .recover(recoveryParams)
            .should.rejectedWith('Error during ICP recovery: Transaction broadcast error: Request failed with status code 500');
    });
    it('should fail to recover txn if balance is low', async () => {
        sinon_1.default.restore();
        setupLowBalanceStubs();
        await icp
            .recover(recoveryParams)
            .should.rejectedWith('Error during ICP recovery: Did not have enough funds to recover');
    });
    it('should fail to recover txn if userKey is not provided', async () => {
        recoveryParams.userKey = '';
        await icp.recover(recoveryParams).should.rejectedWith('Error during ICP recovery: missing userKey');
    });
    it('should fail to recover txn if backupKey is not provided', async () => {
        recoveryParams.backupKey = '';
        await icp.recover(recoveryParams).should.rejectedWith('Error during ICP recovery: missing backupKey');
    });
    it('should fail to recover txn if wallet passphrase is not provided', async () => {
        recoveryParams.walletPassphrase = '';
        await icp.recover(recoveryParams).should.rejectedWith('Error during ICP recovery: missing wallet passphrase');
    });
});
