// @bitgo/sdk-coin-celo/dist/test/unit/celo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const src_1 = require("../../src");
const sdk_core_1 = require("@bitgo/sdk-core");
describe('Celo Gold', function () {
    let bitgo;
    let basecoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('tcelo', src_1.Tcelo.createInstance);
        bitgo.safeRegister('celo', src_1.Celo.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('celo');
    });
    it('should instantiate the coin', function () {
        let localBasecoin = bitgo.coin('tcelo');
        localBasecoin.should.be.an.instanceof(src_1.Tcelo);
        localBasecoin = bitgo.coin('celo');
        localBasecoin.should.be.an.instanceof(src_1.Celo);
    });
    describe('Verify Transaction', function () {
        const address1 = '5Ge59qRnZa8bxyhVFE6BDoY3kuhSrNVETRxXYLty1Hh6XTaf';
        const address2 = '5DiMLZugmcKEH3igPZP367FqummZkWeW5Z6zDCHLfxRjnPXe';
        it('should reject a txPrebuild with more than one recipient', async function () {
            const wallet = new sdk_core_1.Wallet(bitgo, basecoin, {});
            const txParams = {
                recipients: [
                    { amount: '1000000000000', address: address1 },
                    { amount: '2500000000000', address: address2 },
                ],
                wallet: wallet,
                walletPassphrase: 'fakeWalletPassphrase',
            };
            await basecoin
                .verifyTransaction({ txParams })
                .should.be.rejectedWith(`celo doesn't support sending to more than 1 destination address within a single transaction. Try again, using only a single recipient.`);
        });
    });
});
