// @bitgo/abstract-utxo/dist/cjs/src/address/fixedScript.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateAddressWithChainAndIndex = generateAddressWithChainAndIndex;
exports.generateAddress = generateAddress;
exports.assertFixedScriptWalletAddress = assertFixedScriptWalletAddress;
const lodash_1 = __importDefault(require("lodash"));
const sdk_core_1 = require("@bitgo/sdk-core");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
function supportsAddressType(coinName, addressType) {
    return wasm_utxo_1.fixedScriptWallet.supportsScriptType(coinName, addressType);
}
/**
 * Normalize script type aliases. "p2tr" is an alias for "p2trLegacy".
 */
function normalizeScriptType(scriptType) {
    return scriptType === 'p2tr' ? 'p2trLegacy' : scriptType;
}
function generateAddressWithChainAndIndex(coinName, keychains, chain, index, format) {
    // Convert CreateAddressFormat to AddressFormat for wasm-utxo
    // 'base58' -> 'default', 'cashaddr' -> 'cashaddr'
    const wasmFormat = format === 'base58' ? 'default' : format;
    return wasm_utxo_1.fixedScriptWallet.address(keychains, chain, index, coinName, wasmFormat);
}
/**
 * Generate an address for a wallet based on a set of configurations
 * @param params.keychains   {[object]} Array of objects with xpubs
 * @param params.threshold   {number}   Minimum number of signatures
 * @param params.chain       {number}   Derivation chain (see https://github.com/BitGo/unspents/blob/master/src/codes.ts for
 *                                                 the corresponding address type of a given chain code)
 * @param params.index       {number}   Derivation index
 * @param params.segwit      {boolean}  Deprecated
 * @param params.bech32      {boolean}  Deprecated
 * @returns {string} The generated address
 */
function generateAddress(coinName, params) {
    let derivationIndex = 0;
    if (lodash_1.default.isInteger(params.index) && params.index > 0) {
        derivationIndex = params.index;
    }
    const { keychains, chain, segwit = false, bech32 = false } = params;
    let derivationChain = wasm_utxo_1.fixedScriptWallet.ChainCode.value('p2sh', 'external');
    if (lodash_1.default.isNumber(chain) && lodash_1.default.isInteger(chain) && wasm_utxo_1.fixedScriptWallet.ChainCode.is(chain)) {
        derivationChain = chain;
    }
    function convertFlagsToAddressType() {
        if (wasm_utxo_1.fixedScriptWallet.ChainCode.is(chain)) {
            return wasm_utxo_1.fixedScriptWallet.ChainCode.scriptType(chain);
        }
        if (lodash_1.default.isBoolean(segwit) && segwit) {
            return 'p2shP2wsh';
        }
        else if (lodash_1.default.isBoolean(bech32) && bech32) {
            return 'p2wsh';
        }
        else {
            return 'p2sh';
        }
    }
    const addressType = normalizeScriptType(convertFlagsToAddressType());
    if (addressType !== wasm_utxo_1.fixedScriptWallet.ChainCode.scriptType(derivationChain)) {
        throw new sdk_core_1.AddressTypeChainMismatchError(addressType, derivationChain);
    }
    if (!supportsAddressType(coinName, addressType)) {
        switch (addressType) {
            case 'p2sh':
                throw new Error(`internal error: p2sh should always be supported`);
            case 'p2shP2wsh':
                throw new sdk_core_1.P2shP2wshUnsupportedError();
            case 'p2wsh':
                throw new sdk_core_1.P2wshUnsupportedError();
            case 'p2tr':
            case 'p2trLegacy':
                throw new sdk_core_1.P2trUnsupportedError();
            case 'p2trMusig2':
                throw new sdk_core_1.P2trMusig2UnsupportedError();
            default:
                throw new sdk_core_1.UnsupportedAddressTypeError();
        }
    }
    if (!(0, sdk_core_1.isTriple)(keychains)) {
        throw new Error('keychains must be a triple');
    }
    return generateAddressWithChainAndIndex(coinName, keychains.map((k) => k.pub), derivationChain, derivationIndex, params.format);
}
function assertFixedScriptWalletAddress(coinName, { chain, index, keychains, format, address, }) {
    if ((lodash_1.default.isUndefined(chain) && lodash_1.default.isUndefined(index)) || !(lodash_1.default.isFinite(chain) && lodash_1.default.isFinite(index))) {
        throw new sdk_core_1.InvalidAddressDerivationPropertyError(`address validation failure: invalid chain (${chain}) or index (${index})`);
    }
    if (!keychains) {
        throw new Error('missing required param keychains');
    }
    const expectedAddress = generateAddress(coinName, {
        format,
        keychains,
        chain,
        index,
    });
    if (expectedAddress !== address) {
        throw new sdk_core_1.UnexpectedAddressError(`address validation failure: expected ${expectedAddress} but got ${address}`);
    }
}
