// @babylonlabs-io/babylon-proto-ts/dist/generated/babylon/btclightclient/v1/params.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.6.1
//   protoc               unknown
// source: babylon/btclightclient/v1/params.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "babylon.btclightclient.v1";
function createBaseParams() {
    return { insertHeadersAllowList: [] };
}
exports.Params = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.insertHeadersAllowList) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.insertHeadersAllowList.push(reader.string());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            insertHeadersAllowList: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.insertHeadersAllowList)
                ? object.insertHeadersAllowList.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.insertHeadersAllowList) === null || _a === void 0 ? void 0 : _a.length) {
            obj.insertHeadersAllowList = message.insertHeadersAllowList;
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParams();
        message.insertHeadersAllowList = ((_a = object.insertHeadersAllowList) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
