// @bitgo/sdk-coin-tao/dist/test/unit/transactionBuilder/moveStakeBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const sinon_1 = require("sinon");
const moveStakeBuilder_1 = require("../../../src/lib/moveStakeBuilder");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const resources_1 = require("../../resources");
const base_1 = require("./base");
const resources_2 = require("../../../src/resources");
const sdk_core_1 = require("@bitgo/sdk-core");
// Test helper class to access private methods for testing
class TestMoveStakeBuilder extends moveStakeBuilder_1.MoveStakeBuilder {
    setMethodForTesting(method) {
        this._method = method;
    }
}
describe('Tao Move Stake Builder', function () {
    const referenceBlock = '0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d';
    let builder;
    const sender = resources_1.accounts.account1;
    beforeEach(function () {
        const config = (0, base_1.buildTestConfig)();
        const material = utils_1.default.getMaterial(config.network.type);
        builder = new moveStakeBuilder_1.MoveStakeBuilder(config).material(material);
    });
    describe('setter validation', function () {
        it('should validate amount', function () {
            assert_1.default.throws(() => builder.amount('-1'), (e) => e.message === 'Amount must be greater than zero');
            assert_1.default.throws(() => builder.amount('0'), (e) => e.message === 'Amount must be greater than zero');
            should_1.default.doesNotThrow(() => builder.amount('1000'));
            should_1.default.doesNotThrow(() => builder.amount('1'));
        });
        it('should validate addresses', function () {
            const spyValidateAddress = (0, sinon_1.spy)(builder, 'validateAddress');
            assert_1.default.throws(() => builder.originHotkey({ address: 'abc' }), (e) => e.message === `The address 'abc' is not a well-formed dot address`);
            assert_1.default.throws(() => builder.destinationHotkey({ address: 'abc' }), (e) => e.message === `The address 'abc' is not a well-formed dot address`);
            should_1.default.doesNotThrow(() => builder.originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' }));
            should_1.default.doesNotThrow(() => builder.destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' }));
            sinon_1.assert.callCount(spyValidateAddress, 4);
        });
    });
    describe('build move stake transaction', function () {
        it('should build a move stake transaction', async function () {
            builder
                .amount('9007199254740995')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('1')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const tx = await builder.build();
            const txJson = tx.toJson();
            txJson.should.have.properties([
                'id',
                'sender',
                'referenceBlock',
                'blockNumber',
                'genesisHash',
                'nonce',
                'specVersion',
                'transactionVersion',
                'eraPeriod',
                'chainName',
                'tip',
                'originHotkey',
                'destinationHotkey',
                'originNetuid',
                'destinationNetuid',
                'alphaAmount',
            ]);
            txJson.sender.should.equal('5EGoFA95omzemRssELLDjVenNZ68aXyUeqtKQScXSEBvVJkr');
            txJson.originHotkey.should.equal('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT');
            txJson.destinationHotkey.should.equal('5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq');
            txJson.originNetuid.should.equal('1');
            txJson.destinationNetuid.should.equal('1');
            txJson.alphaAmount.should.equal('9007199254740995');
            txJson.blockNumber.should.equal(3933);
            txJson.nonce.should.equal(200);
            txJson.tip.should.equal(0);
            // Verify transaction explanation
            const explanation = tx.explainTransaction();
            explanation.should.have.properties(['outputAmount', 'changeAmount', 'fee']);
        });
        it('should validate required fields', function () {
            assert_1.default.throws(() => builder.validateTransaction({}), (e) => e.message.includes('Transaction validation failed'));
        });
        it('should set and get origin netuid', function () {
            builder.originNetuid('5');
            // We can't directly access private fields, but we can verify through building
            should_1.default.doesNotThrow(() => builder.originNetuid('5'));
        });
        it('should set and get destination netuid', function () {
            builder.destinationNetuid('10');
            // We can't directly access private fields, but we can verify through building
            should_1.default.doesNotThrow(() => builder.destinationNetuid('10'));
        });
        it('should build transaction with different netuids', async function () {
            builder
                .amount('1000000000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('2')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const tx = await builder.build();
            const txJson = tx.toJson();
            txJson.originNetuid.should.equal('1');
            txJson.destinationNetuid.should.equal('2');
            txJson.alphaAmount.should.equal('1000000000000');
        });
    });
    describe('validation', function () {
        it('should validate move stake transaction schema', function () {
            should_1.default.doesNotThrow(() => {
                builder
                    .amount('1000000000000')
                    .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                    .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                    .originNetuid('1')
                    .destinationNetuid('1');
            });
        });
        it('should throw error for invalid amount', function () {
            assert_1.default.throws(() => builder.amount('-100'), (e) => e.message.includes('Amount must be greater than zero'));
        });
        it('should throw error for zero amount', function () {
            assert_1.default.throws(() => builder.amount('0'), (e) => e.message === 'Amount must be greater than zero');
        });
        it('should validate netuid range for origin netuid', function () {
            // Valid netuids
            should_1.default.doesNotThrow(() => builder.originNetuid('0'));
            should_1.default.doesNotThrow(() => builder.originNetuid('64'));
            should_1.default.doesNotThrow(() => builder.originNetuid('128'));
            // Invalid netuids
            assert_1.default.throws(() => builder.originNetuid('-1'), (e) => e.message.includes('Invalid netuid: -1. Must be a non-negative integer.'));
            assert_1.default.throws(() => builder.originNetuid('129'), (e) => e.message.includes('Invalid netuid: 129. Netuid must be between 0 and 128.'));
            assert_1.default.throws(() => builder.originNetuid('abc'), (e) => e.message.includes('Invalid netuid: abc. Must be a non-negative integer.'));
        });
        it('should validate netuid range for destination netuid', function () {
            // Valid netuids
            should_1.default.doesNotThrow(() => builder.destinationNetuid('0'));
            should_1.default.doesNotThrow(() => builder.destinationNetuid('64'));
            should_1.default.doesNotThrow(() => builder.destinationNetuid('128'));
            // Invalid netuids
            assert_1.default.throws(() => builder.destinationNetuid('-1'), (e) => e.message.includes('Invalid netuid: -1. Must be a non-negative integer.'));
            assert_1.default.throws(() => builder.destinationNetuid('129'), (e) => e.message.includes('Invalid netuid: 129. Netuid must be between 0 and 128.'));
            assert_1.default.throws(() => builder.destinationNetuid('invalid'), (e) => e.message.includes('Invalid netuid: invalid. Must be a non-negative integer.'));
        });
    });
    describe('TSS signature integration', function () {
        it('should build a signed move stake transaction with TSS signature', async function () {
            builder
                .amount('9007199254740995')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('1')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' })
                .addSignature({ pub: sender.publicKey }, Buffer.from(resources_1.mockTssSignature, 'hex'));
            const tx = await builder.build();
            const txJson = tx.toJson();
            txJson.alphaAmount.should.equal('9007199254740995');
            txJson.originHotkey.should.equal('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT');
            txJson.destinationHotkey.should.equal('5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq');
            txJson.originNetuid.should.equal('1');
            txJson.destinationNetuid.should.equal('1');
            txJson.sender.should.equal(sender.address);
            txJson.blockNumber.should.equal(3933);
            txJson.referenceBlock.should.equal(referenceBlock);
            txJson.genesisHash.should.equal(resources_1.genesisHash);
            txJson.specVersion.should.equal(Number(resources_2.testnetMaterial.specVersion));
            txJson.nonce.should.equal(200);
            txJson.tip.should.equal(0);
            txJson.transactionVersion.should.equal(Number(resources_2.testnetMaterial.txVersion));
            txJson.chainName.toLowerCase().should.equal(resources_1.chainName);
            txJson.eraPeriod.should.equal(64);
        });
        it('should build an unsigned move stake transaction', async function () {
            builder
                .amount('50000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('2')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const tx = await builder.build();
            const txJson = tx.toJson();
            txJson.alphaAmount.should.equal('50000000');
            txJson.originHotkey.should.equal('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT');
            txJson.destinationHotkey.should.equal('5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq');
            txJson.originNetuid.should.equal('1');
            txJson.destinationNetuid.should.equal('2');
            txJson.sender.should.equal(sender.address);
            txJson.blockNumber.should.equal(3933);
            txJson.referenceBlock.should.equal(referenceBlock);
            txJson.genesisHash.should.equal(resources_1.genesisHash);
            txJson.specVersion.should.equal(Number(resources_2.testnetMaterial.specVersion));
            txJson.nonce.should.equal(200);
            txJson.tip.should.equal(0);
            txJson.transactionVersion.should.equal(Number(resources_2.testnetMaterial.txVersion));
            txJson.chainName.toLowerCase().should.equal(resources_1.chainName);
            txJson.eraPeriod.should.equal(64);
        });
    });
    describe('comprehensive error handling', function () {
        it('should throw error for missing origin hotkey', function () {
            builder
                .amount('1000000000000')
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('1');
            assert_1.default.throws(() => builder.validateTransaction({}), (e) => e.message.includes('Transaction validation failed'));
        });
        it('should throw error for missing destination hotkey', function () {
            builder
                .amount('1000000000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .originNetuid('1')
                .destinationNetuid('1');
            assert_1.default.throws(() => builder.validateTransaction({}), (e) => e.message.includes('Transaction validation failed'));
        });
        it('should throw error for missing origin netuid', function () {
            builder
                .amount('1000000000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .destinationNetuid('1');
            assert_1.default.throws(() => builder.validateTransaction({}), (e) => e.message.includes('Transaction validation failed'));
        });
        it('should throw error for missing destination netuid', function () {
            builder
                .amount('1000000000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1');
            assert_1.default.throws(() => builder.validateTransaction({}), (e) => e.message.includes('Transaction validation failed'));
        });
        it('should throw error for missing amount', function () {
            builder
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('1');
            assert_1.default.throws(() => builder.validateTransaction({}), (e) => e.message.includes('Transaction validation failed'));
        });
        it('should throw error for invalid transaction type in fromImplementation', function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const mockBuilder = new TestMoveStakeBuilder(config).material(material);
            mockBuilder.setMethodForTesting({
                name: 'transferKeepAlive',
                args: { dest: { id: 'test' }, value: '1000' },
                pallet: 'balances',
            });
            assert_1.default.throws(() => {
                // Call the validation logic directly
                if (mockBuilder['_method']?.name !== 'moveStake') {
                    throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${mockBuilder['_method']?.name}. Expected moveStake`);
                }
            }, (e) => e.message.includes('Invalid Transaction Type: transferKeepAlive. Expected moveStake'));
        });
        it('should handle malformed raw transaction', function () {
            assert_1.default.throws(() => builder.from('invalid_hex_data'), (e) => e.message !== undefined);
        });
    });
    describe('boundary value and edge case tests', function () {
        it('should handle very large amounts', function () {
            const largeAmount = '999999999999999999999999999999';
            should_1.default.doesNotThrow(() => builder.amount(largeAmount));
        });
        it('should handle same origin and destination hotkeys', function () {
            const sameAddress = '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT';
            should_1.default.doesNotThrow(() => {
                builder.originHotkey({ address: sameAddress }).destinationHotkey({ address: sameAddress });
            });
        });
        it('should handle same origin and destination netuids', function () {
            should_1.default.doesNotThrow(() => {
                builder.originNetuid('5').destinationNetuid('5');
            });
        });
        it('should validate various address formats', function () {
            const invalidAddresses = [
                '',
                '123',
                'invalid_address_format',
                '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT_invalid',
            ];
            invalidAddresses.forEach((address) => {
                assert_1.default.throws(() => builder.originHotkey({ address }), (e) => e.message.includes('is not a well-formed dot address'));
                assert_1.default.throws(() => builder.destinationHotkey({ address }), (e) => e.message.includes('is not a well-formed dot address'));
            });
        });
        it('should handle boundary netuid values', function () {
            should_1.default.doesNotThrow(() => builder.originNetuid('0'));
            should_1.default.doesNotThrow(() => builder.originNetuid('128'));
            should_1.default.doesNotThrow(() => builder.destinationNetuid('0'));
            should_1.default.doesNotThrow(() => builder.destinationNetuid('128'));
            assert_1.default.throws(() => builder.originNetuid('-1'), (e) => e.message.includes('Invalid netuid: -1. Must be a non-negative integer.'));
            assert_1.default.throws(() => builder.destinationNetuid('129'), (e) => e.message.includes('Invalid netuid: 129. Netuid must be between 0 and 128.'));
        });
    });
    describe('transaction explanation validation', function () {
        it('should provide correct explanation with different subnet tokens', async function () {
            builder
                .amount('1000000000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('2')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const tx = await builder.build();
            const explanation = tx.explainTransaction();
            explanation.should.have.properties(['outputAmount', 'changeAmount', 'fee', 'type']);
            explanation.outputAmount.should.equal('1000000000000');
            explanation.changeAmount.should.equal('0');
            explanation.fee.should.have.properties(['fee', 'type']);
            explanation.fee.type.should.equal('tip');
        });
        it('should handle explanation with zero tip', async function () {
            builder
                .amount('500000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('1')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const tx = await builder.build();
            const explanation = tx.explainTransaction();
            explanation.fee.fee.should.equal('0');
            explanation.outputAmount.should.equal('500000000');
        });
    });
    describe('fromImplementation stages validation', function () {
        it('should call super.fromImplementation before validation to populate _method', async function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const validBuilder = new moveStakeBuilder_1.MoveStakeBuilder(config).material(material);
            validBuilder
                .amount('1000000000000')
                .originHotkey({ address: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT' })
                .destinationHotkey({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .originNetuid('1')
                .destinationNetuid('2')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const validTx = await validBuilder.build();
            const rawTxHex = validTx.toBroadcastFormat();
            const newBuilder = new moveStakeBuilder_1.MoveStakeBuilder(config).material(material);
            should_1.default.doesNotThrow(() => {
                newBuilder.from(rawTxHex);
            });
            const builderMethod = newBuilder._method;
            builderMethod.should.not.be.undefined();
            builderMethod.name.should.equal('moveStake');
            builderMethod.args.should.have.properties([
                'originHotkey',
                'destinationHotkey',
                'originNetuid',
                'destinationNetuid',
                'alphaAmount',
            ]);
            builderMethod.args.alphaAmount.should.equal('1000000000000');
            builderMethod.args.originHotkey.should.equal('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT');
            builderMethod.args.destinationHotkey.should.equal('5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq');
            builderMethod.args.originNetuid.should.equal('1');
            builderMethod.args.destinationNetuid.should.equal('2');
        });
        it('should throw error if _method is not populated before validation', function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const mockBuilder = new TestMoveStakeBuilder(config).material(material);
            assert_1.default.throws(() => {
                if (mockBuilder['_method']?.name !== 'moveStake') {
                    throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${mockBuilder['_method']?.name}. Expected moveStake`);
                }
            }, (e) => e.message.includes('Invalid Transaction Type: undefined. Expected moveStake'));
        });
        it('should properly validate transaction type after super.fromImplementation', function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const mockBuilder = new TestMoveStakeBuilder(config).material(material);
            mockBuilder.setMethodForTesting({
                name: 'transferKeepAlive',
                args: { dest: { id: 'test' }, value: '1000' },
                pallet: 'balances',
            });
            assert_1.default.throws(() => {
                if (mockBuilder['_method']?.name !== 'moveStake') {
                    throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${mockBuilder['_method']?.name}. Expected moveStake`);
                }
            }, (e) => e.message.includes('Invalid Transaction Type: transferKeepAlive. Expected moveStake'));
        });
        it('should successfully parse and validate correct moveStake transaction', function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const mockBuilder = new TestMoveStakeBuilder(config).material(material);
            mockBuilder.setMethodForTesting({
                name: 'moveStake',
                args: {
                    originHotkey: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT',
                    destinationHotkey: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq',
                    originNetuid: '1',
                    destinationNetuid: '2',
                    alphaAmount: '1000000000000',
                },
                pallet: 'subtensorModule',
            });
            should_1.default.doesNotThrow(() => {
                if (mockBuilder['_method']?.name !== 'moveStake') {
                    throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${mockBuilder['_method']?.name}. Expected moveStake`);
                }
            });
        });
        it('should fail validation when parsing wrong transaction type (transferStake instead of moveStake)', function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const moveStakeBuilder = new moveStakeBuilder_1.MoveStakeBuilder(config).material(material);
            assert_1.default.throws(() => {
                moveStakeBuilder.from(resources_1.rawTx.transferStake.signed);
            }, (e) => e.message.includes('Invalid Transaction Type: transferStake. Expected moveStake'));
        });
        it('should verify _method is properly populated after super.fromImplementation with wrong transaction type', function () {
            const config = (0, base_1.buildTestConfig)();
            const material = utils_1.default.getMaterial(config.network.type);
            const testBuilder = new TestMoveStakeBuilder(config).material(material);
            try {
                testBuilder.from(resources_1.rawTx.transferStake.signed);
            }
            catch (error) {
                const method = testBuilder._method;
                method.should.not.be.undefined();
                method.name.should.equal('transferStake'); // This proves super.fromImplementation was called
                method.should.have.property('args');
                method.should.have.property('pallet');
                error.message.should.containEql('Invalid Transaction Type: transferStake. Expected moveStake');
            }
        });
    });
});
