// @bitgo/sdk-coin-tao/dist/test/unit/transactionBuilder/tokenTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const sinon_1 = require("sinon");
const tokenTransferBuilder_1 = require("../../../src/lib/tokenTransferBuilder");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const resources_1 = require("../../resources");
const base_1 = require("./base");
describe('Tao Token Transfer Builder', function () {
    const referenceBlock = '0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d';
    let builder;
    const sender = resources_1.accounts.account1;
    beforeEach(function () {
        const config = (0, base_1.buildTestConfig)();
        const material = utils_1.default.getMaterial(config.network.type);
        builder = new tokenTransferBuilder_1.TokenTransferBuilder(config).material(material);
    });
    describe('setter validation', function () {
        it('should validate amount', function () {
            const spyValidateValue = (0, sinon_1.spy)(builder, 'validateValue');
            assert_1.default.throws(() => builder.amount('-1'), (e) => e.message === 'Value cannot be less than zero');
            should_1.default.doesNotThrow(() => builder.amount('1000'));
            sinon_1.assert.calledTwice(spyValidateValue);
        });
        it('should validate address', function () {
            const spyValidateAddress = (0, sinon_1.spy)(builder, 'validateAddress');
            assert_1.default.throws(() => builder.hotkey('abc'), (e) => e.message === `The address 'abc' is not a well-formed dot address`);
            assert_1.default.throws(() => builder.destinationColdkey('abc'), (e) => e.message === `The address 'abc' is not a well-formed dot address`);
            should_1.default.doesNotThrow(() => builder.hotkey('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT'));
            should_1.default.doesNotThrow(() => builder.destinationColdkey('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT'));
            sinon_1.assert.callCount(spyValidateAddress, 4);
        });
        it('should validate netuid', function () {
            const spyValidateAddress = (0, sinon_1.spy)(builder, 'validateNetuid');
            assert_1.default.throws(() => builder.destinationNetuid('abc'), (e) => e.message === `Cannot convert abc to a BigInt`);
            assert_1.default.throws(() => builder.destinationNetuid('-10'), (e) => e.message === `The netuid '-10' is not a valid netuid`);
            should_1.default.doesNotThrow(() => builder.destinationNetuid('36'));
            assert_1.default.throws(() => builder.originNetuid('abc'), (e) => e.message === `Cannot convert abc to a BigInt`);
            assert_1.default.throws(() => builder.destinationNetuid('-1'), (e) => e.message === `The netuid '-1' is not a valid netuid`);
            should_1.default.doesNotThrow(() => builder.originNetuid('64'));
            sinon_1.assert.callCount(spyValidateAddress, 6);
        });
    });
    describe('build transfer stake transaction', function () {
        it('should build a transfer stake transaction', async function () {
            builder
                .amount('9007199254740995')
                .destinationColdkey('5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq')
                .hotkey('5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT')
                .originNetuid('12')
                .destinationNetuid('10')
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock(referenceBlock)
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' })
                .addSignature({ pub: sender.publicKey }, Buffer.from(resources_1.mockTssSignature, 'hex'));
            const tx = await builder.build();
            const serializedTx = tx.toBroadcastFormat();
            serializedTx.should.equal(resources_1.rawTx.transferStake.signed);
            tx.toJson().should.deepEqual({
                id: '0xe5932b507a53d351aabd520487ba55ba833e0968ad18642a28f54dabeb7abf2f',
                sender: '5EGoFA95omzemRssELLDjVenNZ68aXyUeqtKQScXSEBvVJkr',
                referenceBlock: '0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d',
                blockNumber: 3933,
                genesisHash: '0x8f9cf856bf558a14440e75569c9e58594757048d7b3a84b5d25f6bd978263105',
                nonce: 200,
                specVersion: 224,
                transactionVersion: 1,
                eraPeriod: 64,
                chainName: 'Bittensor',
                tip: 0,
                destinationColdkey: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq',
                hotkey: '5FCPTnjevGqAuTttetBy4a24Ej3pH9fiQ8fmvP1ZkrVsLUoT',
                originNetuid: '12',
                destinationNetuid: '10',
                alphaAmount: '9007199254740995',
            });
        });
        it('should re-build from raw signed tx', async function () {
            builder.from(resources_1.rawTx.transferStake.signed);
            builder.validity({ firstValid: 3933, maxDuration: 64 }).referenceBlock(referenceBlock);
            const tx = await builder.build();
            const serializedTx = tx.toBroadcastFormat();
            serializedTx.should.equal(resources_1.rawTx.transferStake.signed);
        });
    });
});
