// @bitgo/sdk-coin-evm/dist/src/ethLikeErc20Token.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthLikeErc20Token = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
const assert_1 = __importDefault(require("assert"));
class EthLikeErc20Token extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig, coinNames) {
        super(bitgo, tokenConfig, coinNames);
        this.coinNames = coinNames;
    }
    static createTokenConstructor(config, coinNames) {
        return (bitgo) => new this(bitgo, config, coinNames);
    }
    static createTokenConstructors(coinNames) {
        return super.createTokenConstructors(coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    supportsTss() {
        return true;
    }
    async recoveryBlockchainExplorerQuery(query) {
        const family = this.getFamily();
        const evmConfig = sdk_core_1.common.Environments[this.bitgo.getEnv()].evm;
        (0, assert_1.default)(evmConfig && this.getFamily() in evmConfig, `env config is missing for ${this.getFamily()} in ${this.bitgo.getEnv()}`);
        const explorerUrl = evmConfig[family].baseUrl;
        const apiToken = evmConfig[family].apiToken;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    //TODO: implement a way to return the coin family name or coin name instead of standard ERC20 Token.
    getFullName() {
        return 'ERC20 Token';
    }
}
exports.EthLikeErc20Token = EthLikeErc20Token;
