// @bitgo/sdk-coin-evm/dist/src/evmCoin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvmCoin = void 0;
/**
 * @prettier
 */
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
const utils_1 = require("./lib/utils");
const assert_1 = __importDefault(require("assert"));
class EvmCoin extends abstract_eth_1.AbstractEthLikeNewCoins {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
    }
    static createInstance(bitgo, staticsCoin) {
        return new EvmCoin(bitgo, staticsCoin);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc */
    supportsTss() {
        return this.staticsCoin?.features.includes(statics_1.CoinFeature.TSS) ?? false;
    }
    /** inherited doc */
    getDefaultMultisigType() {
        return this.staticsCoin?.features.includes(statics_1.CoinFeature.TSS) ? sdk_core_1.multisigTypes.tss : sdk_core_1.multisigTypes.onchain;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    async buildUnsignedSweepTxnTSS(params) {
        if (this.staticsCoin?.features.includes(statics_1.CoinFeature.MPCV2)) {
            return this.buildUnsignedSweepTxnMPCv2(params);
        }
        return super.buildUnsignedSweepTxnTSS(params);
    }
    /**
     * Make a query to chain explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @param {string} apiKey optional API key to use for the query falls back to EVM config if not provided
     * @returns {Promise<Object>} response from chain explorer
     */
    async recoveryBlockchainExplorerQuery(query, apiKey) {
        const evmConfig = sdk_core_1.common.Environments[this.bitgo.getEnv()].evm;
        (0, assert_1.default)(evmConfig && this.getFamily() in evmConfig, `env config is missing for ${this.getFamily()} in ${this.bitgo.getEnv()}`);
        const apiToken = apiKey || evmConfig[this.getFamily()].apiToken;
        const explorerUrl = evmConfig[this.getFamily()].baseUrl;
        switch (this.getFamily()) {
            case statics_1.CoinFamily.HBAREVM:
                (0, assert_1.default)(evmConfig[this.getFamily()].rpcUrl, `rpc url config is missing for ${this.getFamily()} in ${this.bitgo.getEnv()}`);
                const rpcUrl = evmConfig[this.getFamily()].rpcUrl;
                return await (0, utils_1.recovery_HBAREVM_BlockchainExplorerQuery)(query, rpcUrl, explorerUrl, apiToken);
            default:
                return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
        }
    }
    /** @inheritDoc */
    async verifyTssTransaction(params) {
        const supportsEIP1559 = this.staticsCoin?.features?.includes(statics_1.CoinFeature.EIP1559);
        if (supportsEIP1559) {
            return await super.verifyTssTransaction(params);
        }
        else {
            return await this.verifyLegacyTssTransaction(params);
        }
    }
    /**
     * Verifies legacy (non-EIP-1559) TSS transactions with basic validation.
     */
    async verifyLegacyTssTransaction(params) {
        const { txParams, txPrebuild, wallet } = params;
        // Basic validation for legacy transactions only
        if (!txParams?.recipients &&
            !(txParams.prebuildTx?.consolidateId ||
                (txParams.type &&
                    ['acceleration', 'fillNonce', 'transferToken', 'tokenApproval', 'bridgeFunds'].includes(txParams.type)))) {
            throw new Error(`missing txParams`);
        }
        if (!wallet || !txPrebuild) {
            throw new Error(`missing params`);
        }
        if (txParams.hop && txParams.recipients && txParams.recipients.length > 1) {
            throw new Error(`tx cannot be both a batch and hop transaction`);
        }
        // If validation passes, consider it verified
        return true;
    }
    /** @inheritDoc */
    isValidAddress(address, isAlternateAddress) {
        if (isAlternateAddress && this.getFamily() === statics_1.CoinFamily.HBAREVM) {
            const { valid } = (0, utils_1.validateHederaAccountId)(address);
            return valid;
        }
        return super.isValidAddress(address);
    }
}
exports.EvmCoin = EvmCoin;
