// @bitgo/sdk-coin-sol/dist/cjs/src/lib/closeAtaBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloseAtaBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const assert_1 = __importDefault(require("assert"));
const constants_1 = require("./constants");
const transaction_1 = require("./transaction");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = require("./utils");
class CloseAtaBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.CloseAssociatedTokenAccount;
    }
    accountAddress(accountAddress) {
        (0, utils_1.validateAddress)(accountAddress, 'accountAddress');
        this._accountAddress = accountAddress;
        return this;
    }
    destinationAddress(destinationAddress) {
        (0, utils_1.validateAddress)(destinationAddress, 'destinationAddress');
        this._destinationAddress = destinationAddress;
        return this;
    }
    authorityAddress(authorityAddress) {
        (0, utils_1.validateAddress)(authorityAddress, 'authorityAddress');
        this._authorityAddress = authorityAddress;
        return this;
    }
    /** @inheritDoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.CloseAssociatedTokenAccount) {
                const ataCloseInstruction = instruction;
                this.accountAddress(ataCloseInstruction.params.accountAddress);
                this.destinationAddress(ataCloseInstruction.params.destinationAddress);
                this.authorityAddress(ataCloseInstruction.params.authorityAddress);
            }
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        (0, assert_1.default)(this._accountAddress, 'Account Address must be set before building the transaction');
        (0, assert_1.default)(this._destinationAddress, 'Destination Address must be set before building the transaction');
        (0, assert_1.default)(this._authorityAddress, 'Authority Address must be set before building the transaction');
        if (this._accountAddress === this._destinationAddress) {
            throw new sdk_core_1.BuildTransactionError('Account address to close cannot be the same as the destination address');
        }
        const closeAssociatedTokenAccountData = {
            type: constants_1.InstructionBuilderTypes.CloseAssociatedTokenAccount,
            params: {
                accountAddress: this._accountAddress,
                destinationAddress: this._destinationAddress,
                authorityAddress: this._authorityAddress,
            },
        };
        this._instructionsData = [closeAssociatedTokenAccountData];
        return await super.buildImplementation();
    }
}
exports.CloseAtaBuilder = CloseAtaBuilder;
