// @bitgo/sdk-coin-sol/dist/cjs/src/lib/customInstructionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomInstructionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const web3_js_1 = require("@solana/web3.js");
const transaction_1 = require("./transaction");
const transactionBuilder_1 = require("./transactionBuilder");
const constants_1 = require("./constants");
const utils_1 = require("./utils");
const assert_1 = __importDefault(require("assert"));
/**
 * Transaction builder for custom Solana instructions.
 * Allows building transactions with any set of raw Solana instructions.
 */
class CustomInstructionBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._customInstructions = [];
    }
    get transactionType() {
        return sdk_core_1.TransactionType.CustomTx;
    }
    /**
     * Initialize the builder from an existing transaction
     */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.CustomInstruction) {
                const customInstruction = instruction;
                this.addCustomInstruction(customInstruction.params);
            }
            else if (instruction.type === constants_1.InstructionBuilderTypes.VersionedCustomInstruction) {
                const versionedCustomInstruction = instruction;
                this.addCustomInstruction(versionedCustomInstruction.params);
            }
        }
    }
    /**
     * Add a custom instruction to the transaction
     * @param instruction - The custom instruction to add
     * @returns This builder instance
     */
    addCustomInstruction(instruction) {
        this.validateInstruction(instruction);
        if ((0, utils_1.isSolLegacyInstruction)(instruction)) {
            const customInstruction = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: instruction,
            };
            this._customInstructions.push(customInstruction);
        }
        else {
            const versionedCustomInstruction = {
                type: constants_1.InstructionBuilderTypes.VersionedCustomInstruction,
                params: instruction,
            };
            this._customInstructions.push(versionedCustomInstruction);
        }
        return this;
    }
    /**
     * Add multiple custom instructions to the transaction
     * @param instructions - Array of custom instructions to add
     * @returns This builder instance
     */
    addCustomInstructions(instructions) {
        if (!Array.isArray(instructions)) {
            throw new sdk_core_1.BuildTransactionError('Instructions must be an array');
        }
        for (const instruction of instructions) {
            this.addCustomInstruction(instruction);
        }
        return this;
    }
    /**
     * Build transaction from deconstructed VersionedTransaction data
     * @param data - VersionedTransactionData containing instructions, ALTs, and account keys
     * @returns This builder instance
     */
    fromVersionedTransactionData(data) {
        try {
            if (!data || typeof data !== 'object') {
                throw new sdk_core_1.BuildTransactionError('VersionedTransactionData must be a valid object');
            }
            if (!Array.isArray(data.versionedInstructions) || data.versionedInstructions.length === 0) {
                throw new sdk_core_1.BuildTransactionError('versionedInstructions must be a non-empty array');
            }
            if (!Array.isArray(data.addressLookupTables)) {
                throw new sdk_core_1.BuildTransactionError('addressLookupTables must be an array');
            }
            if (!Array.isArray(data.staticAccountKeys) || data.staticAccountKeys.length === 0) {
                throw new sdk_core_1.BuildTransactionError('staticAccountKeys must be a non-empty array');
            }
            if (!data.messageHeader || typeof data.messageHeader !== 'object') {
                throw new sdk_core_1.BuildTransactionError('messageHeader must be a valid object');
            }
            if (typeof data.messageHeader.numRequiredSignatures !== 'number' ||
                data.messageHeader.numRequiredSignatures < 0) {
                throw new sdk_core_1.BuildTransactionError('messageHeader.numRequiredSignatures must be a non-negative number');
            }
            if (typeof data.messageHeader.numReadonlySignedAccounts !== 'number' ||
                data.messageHeader.numReadonlySignedAccounts < 0) {
                throw new sdk_core_1.BuildTransactionError('messageHeader.numReadonlySignedAccounts must be a non-negative number');
            }
            if (typeof data.messageHeader.numReadonlyUnsignedAccounts !== 'number' ||
                data.messageHeader.numReadonlyUnsignedAccounts < 0) {
                throw new sdk_core_1.BuildTransactionError('messageHeader.numReadonlyUnsignedAccounts must be a non-negative number');
            }
            let processedData = data;
            if (this._nonceInfo && this._nonceInfo.params) {
                processedData = this.injectNonceAdvanceInstruction(data);
            }
            this.addCustomInstructions(processedData.versionedInstructions);
            if (!this._transaction) {
                this._transaction = new transaction_1.Transaction(this._coinConfig);
            }
            this._transaction.setVersionedTransactionData(processedData);
            this._transaction.setTransactionType(sdk_core_1.TransactionType.CustomTx);
            if (!this._sender && processedData.staticAccountKeys.length > 0) {
                this._sender = processedData.staticAccountKeys[0];
            }
            return this;
        }
        catch (error) {
            if (error instanceof sdk_core_1.BuildTransactionError) {
                throw error;
            }
            throw new sdk_core_1.BuildTransactionError(`Failed to process versioned transaction data: ${error.message}`);
        }
    }
    /**
     * Inject nonce advance instruction into versioned transaction data for durable nonce support.
     * Reorders accounts so signers appear first (required by Solana MessageV0 format).
     * @param data - Original versioned transaction data
     * @returns Modified versioned transaction data with nonce advance instruction
     * @private
     */
    injectNonceAdvanceInstruction(data) {
        const { walletNonceAddress, authWalletAddress } = this._nonceInfo.params;
        const SYSTEM_PROGRAM = web3_js_1.SystemProgram.programId.toBase58();
        const SYSVAR_RECENT_BLOCKHASHES = web3_js_1.SYSVAR_RECENT_BLOCKHASHES_PUBKEY.toBase58();
        const numSigners = data.messageHeader.numRequiredSignatures;
        const originalSigners = data.staticAccountKeys.slice(0, numSigners);
        const originalNonSigners = data.staticAccountKeys.slice(numSigners);
        if (!originalSigners.includes(authWalletAddress)) {
            originalSigners.push(authWalletAddress);
        }
        const nonSigners = [...originalNonSigners];
        const allKeys = [...originalSigners, ...originalNonSigners];
        if (!allKeys.includes(SYSTEM_PROGRAM))
            nonSigners.push(SYSTEM_PROGRAM);
        if (!allKeys.includes(walletNonceAddress))
            nonSigners.push(walletNonceAddress);
        if (!allKeys.includes(SYSVAR_RECENT_BLOCKHASHES))
            nonSigners.push(SYSVAR_RECENT_BLOCKHASHES);
        const newStaticAccountKeys = [...originalSigners, ...nonSigners];
        const nonceAdvanceInstruction = {
            programIdIndex: newStaticAccountKeys.indexOf(SYSTEM_PROGRAM),
            accountKeyIndexes: [
                newStaticAccountKeys.indexOf(walletNonceAddress),
                newStaticAccountKeys.indexOf(SYSVAR_RECENT_BLOCKHASHES),
                newStaticAccountKeys.indexOf(authWalletAddress),
            ],
            data: '6vx8P', // SystemProgram AdvanceNonceAccount (0x04000000) in base58
        };
        const indexMap = new Map(data.staticAccountKeys.map((key, oldIdx) => [oldIdx, newStaticAccountKeys.indexOf(key)]));
        const remappedInstructions = data.versionedInstructions.map((inst) => ({
            programIdIndex: indexMap.get(inst.programIdIndex) ?? inst.programIdIndex,
            accountKeyIndexes: inst.accountKeyIndexes.map((idx) => indexMap.get(idx) ?? idx),
            data: inst.data,
        }));
        return {
            ...data,
            versionedInstructions: [nonceAdvanceInstruction, ...remappedInstructions],
            staticAccountKeys: newStaticAccountKeys,
            messageHeader: {
                ...data.messageHeader,
                numRequiredSignatures: originalSigners.length,
            },
        };
    }
    /**
     * Clear all custom instructions and versioned transaction data
     * @returns This builder instance
     */
    clearInstructions() {
        this._customInstructions = [];
        if (this._transaction) {
            this._transaction.setVersionedTransactionData(undefined);
        }
        return this;
    }
    /**
     * Get the current custom instructions
     * @returns Array of custom instructions
     */
    getInstructions() {
        return [...this._customInstructions];
    }
    /**
     * Validate custom instruction format
     * @param instruction - The instruction to validate
     */
    validateInstruction(instruction) {
        if (!instruction) {
            throw new sdk_core_1.BuildTransactionError('Instruction cannot be null or undefined');
        }
        if ((0, utils_1.isSolLegacyInstruction)(instruction)) {
            this.validateSolInstruction(instruction);
        }
        else {
            this.validateVersionedInstruction(instruction);
        }
    }
    /**
     * Validate traditional SolInstruction format
     * @param instruction - The traditional instruction to validate
     */
    validateSolInstruction(instruction) {
        if (!instruction.programId || typeof instruction.programId !== 'string') {
            throw new sdk_core_1.BuildTransactionError('Instruction must have a valid programId string');
        }
        // Validate that programId is a valid Solana public key
        try {
            new web3_js_1.PublicKey(instruction.programId);
        }
        catch (error) {
            throw new sdk_core_1.BuildTransactionError('Invalid programId format');
        }
        if (!instruction.keys || !Array.isArray(instruction.keys)) {
            throw new sdk_core_1.BuildTransactionError('Instruction must have valid keys array');
        }
        // Validate each key
        for (const key of instruction.keys) {
            if (!key.pubkey || typeof key.pubkey !== 'string') {
                throw new sdk_core_1.BuildTransactionError('Each key must have a valid pubkey string');
            }
            try {
                new web3_js_1.PublicKey(key.pubkey);
            }
            catch (error) {
                throw new sdk_core_1.BuildTransactionError('Invalid pubkey format in keys');
            }
            if (typeof key.isSigner !== 'boolean') {
                throw new sdk_core_1.BuildTransactionError('Each key must have a boolean isSigner field');
            }
            if (typeof key.isWritable !== 'boolean') {
                throw new sdk_core_1.BuildTransactionError('Each key must have a boolean isWritable field');
            }
        }
        if (instruction.data === undefined || typeof instruction.data !== 'string') {
            throw new sdk_core_1.BuildTransactionError('Instruction must have valid data string');
        }
    }
    /**
     * Validate versioned instruction format
     * @param instruction - The versioned instruction to validate
     */
    validateVersionedInstruction(instruction) {
        if (typeof instruction.programIdIndex !== 'number' || instruction.programIdIndex < 0) {
            throw new sdk_core_1.BuildTransactionError('Versioned instruction must have a valid programIdIndex number');
        }
        if (!instruction.accountKeyIndexes || !Array.isArray(instruction.accountKeyIndexes)) {
            throw new sdk_core_1.BuildTransactionError('Versioned instruction must have valid accountKeyIndexes array');
        }
        // Validate each account key index
        for (const index of instruction.accountKeyIndexes) {
            if (typeof index !== 'number' || index < 0) {
                throw new sdk_core_1.BuildTransactionError('Each accountKeyIndex must be a non-negative number');
            }
        }
        if (instruction.data === undefined || typeof instruction.data !== 'string') {
            throw new sdk_core_1.BuildTransactionError('Versioned instruction must have valid data string');
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        (0, assert_1.default)(this._customInstructions.length > 0, 'At least one custom instruction must be specified');
        // Set the instructions data to our custom instructions
        this._instructionsData = [...this._customInstructions];
        return await super.buildImplementation();
    }
}
exports.CustomInstructionBuilder = CustomInstructionBuilder;
