// @bitgo/sdk-coin-sol/dist/cjs/src/lib/instructionParamsFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.instructionParamsFactory = instructionParamsFactory;
exports.findTokenName = findTokenName;
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const public_types_1 = require("@bitgo/public-types");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const assert_1 = __importDefault(require("assert"));
const constants_1 = require("./constants");
const utils_1 = require("./utils");
const jitoStakePoolOperations_1 = require("./jitoStakePoolOperations");
/**
 * Construct instructions params from Solana instructions
 *
 * @param {TransactionType} type - the transaction type
 * @param {TransactionInstruction[]} instructions - solana instructions
 * @returns {InstructionParams[]} An array containing instruction params
 */
function instructionParamsFactory(type, instructions, coinName, instructionMetadata, _useTokenAddressTokenName) {
    switch (type) {
        case sdk_core_1.TransactionType.WalletInitialization:
            return parseWalletInitInstructions(instructions);
        case sdk_core_1.TransactionType.Send:
            return parseSendInstructions(instructions, instructionMetadata, _useTokenAddressTokenName);
        case sdk_core_1.TransactionType.StakingActivate:
            return parseStakingActivateInstructions(instructions);
        case sdk_core_1.TransactionType.StakingDeactivate:
            return parseStakingDeactivateInstructions(instructions, coinName);
        case sdk_core_1.TransactionType.StakingWithdraw:
            return parseStakingWithdrawInstructions(instructions);
        case sdk_core_1.TransactionType.AssociatedTokenAccountInitialization:
            return parseAtaInitInstructions(instructions, instructionMetadata, _useTokenAddressTokenName);
        case sdk_core_1.TransactionType.CloseAssociatedTokenAccount:
            return parseAtaCloseInstructions(instructions);
        case sdk_core_1.TransactionType.StakingAuthorize:
            return parseStakingAuthorizeInstructions(instructions);
        case sdk_core_1.TransactionType.StakingAuthorizeRaw:
            return parseStakingAuthorizeRawInstructions(instructions);
        case sdk_core_1.TransactionType.StakingDelegate:
            return parseStakingDelegateInstructions(instructions);
        case sdk_core_1.TransactionType.CustomTx:
            return parseCustomInstructions(instructions, instructionMetadata);
        default:
            throw new sdk_core_1.NotSupported('Invalid transaction, transaction type not supported: ' + type);
    }
}
/**
 * Parses Solana instructions to Wallet initialization tx instructions params
 *
 * @param {TransactionInstruction[]} instructions - containing create and initialize nonce solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for Wallet initialization tx
 */
function parseWalletInitInstructions(instructions) {
    const instructionData = [];
    const createInstruction = web3_js_1.SystemInstruction.decodeCreateAccount(instructions[constants_1.walletInitInstructionIndexes.Create]);
    const nonceInitInstruction = web3_js_1.SystemInstruction.decodeNonceInitialize(instructions[constants_1.walletInitInstructionIndexes.InitializeNonceAccount]);
    const walletInit = {
        type: constants_1.InstructionBuilderTypes.CreateNonceAccount,
        params: {
            fromAddress: createInstruction.fromPubkey.toString(),
            nonceAddress: nonceInitInstruction.noncePubkey.toString(),
            authAddress: nonceInitInstruction.authorizedPubkey.toString(),
            amount: createInstruction.lamports.toString(),
        },
    };
    instructionData.push(walletInit);
    const memo = getMemo(instructions, constants_1.walletInitInstructionIndexes);
    if (memo) {
        instructionData.push(memo);
    }
    return instructionData;
}
/**
 * Parses Solana instructions to Send tx instructions params
 * Only supports Memo, Transfer and Advance Nonce Solana instructions
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for Send tx
 */
function parseSendInstructions(instructions, instructionMetadata, _useTokenAddressTokenName) {
    const instructionData = [];
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.Memo:
                const memo = { type: constants_1.InstructionBuilderTypes.Memo, params: { memo: instruction.data.toString() } };
                instructionData.push(memo);
                break;
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.Transfer:
                const transferInstruction = web3_js_1.SystemInstruction.decodeTransfer(instruction);
                const transfer = {
                    type: constants_1.InstructionBuilderTypes.Transfer,
                    params: {
                        fromAddress: transferInstruction.fromPubkey.toString(),
                        toAddress: transferInstruction.toPubkey.toString(),
                        amount: transferInstruction.lamports.toString(),
                    },
                };
                instructionData.push(transfer);
                break;
            case constants_1.ValidInstructionTypesEnum.TokenTransfer:
                let tokenTransferInstruction;
                if (instruction.programId.toString() !== spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
                    tokenTransferInstruction = (0, spl_token_1.decodeTransferCheckedInstruction)(instruction);
                }
                else {
                    tokenTransferInstruction = (0, spl_token_1.decodeTransferCheckedInstruction)(instruction, spl_token_1.TOKEN_2022_PROGRAM_ID);
                }
                const tokenAddress = tokenTransferInstruction.keys.mint.pubkey.toString();
                const tokenName = findTokenName(tokenAddress, instructionMetadata, _useTokenAddressTokenName);
                let programIDForTokenTransfer;
                if (instruction.programId) {
                    programIDForTokenTransfer = instruction.programId.toString();
                }
                const tokenTransfer = {
                    type: constants_1.InstructionBuilderTypes.TokenTransfer,
                    params: {
                        fromAddress: tokenTransferInstruction.keys.owner.pubkey.toString(),
                        toAddress: tokenTransferInstruction.keys.destination.pubkey.toString(),
                        amount: tokenTransferInstruction.data.amount.toString(),
                        tokenName,
                        sourceAddress: tokenTransferInstruction.keys.source.pubkey.toString(),
                        tokenAddress: tokenAddress,
                        programId: programIDForTokenTransfer,
                        decimalPlaces: tokenTransferInstruction.data.decimals,
                    },
                };
                instructionData.push(tokenTransfer);
                break;
            case constants_1.ValidInstructionTypesEnum.Approve:
                const programId = instruction.programId.equals(spl_token_1.TOKEN_2022_PROGRAM_ID) ? spl_token_1.TOKEN_2022_PROGRAM_ID : undefined;
                const approveInstruction = (0, spl_token_1.decodeApproveInstruction)(instruction, programId);
                const approve = {
                    type: constants_1.InstructionBuilderTypes.Approve,
                    params: {
                        accountAddress: approveInstruction.keys.account.toString(),
                        delegateAddress: approveInstruction.keys.delegate.toString(),
                        ownerAddress: approveInstruction.keys.owner.toString(),
                        amount: approveInstruction.data.amount.toString(),
                        programId: programId && programId.toString(),
                    },
                };
                instructionData.push(approve);
                break;
            case constants_1.ValidInstructionTypesEnum.InitializeAssociatedTokenAccount:
                const mintAddress = instruction.keys[ataInitInstructionKeysIndexes.MintAddress].pubkey.toString();
                const mintTokenName = findTokenName(mintAddress, instructionMetadata, _useTokenAddressTokenName);
                let programID;
                if (instruction.programId) {
                    programID = instruction.programId.toString();
                }
                const ataInit = {
                    type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                    params: {
                        mintAddress,
                        ataAddress: instruction.keys[ataInitInstructionKeysIndexes.ATAAddress].pubkey.toString(),
                        ownerAddress: instruction.keys[ataInitInstructionKeysIndexes.OwnerAddress].pubkey.toString(),
                        payerAddress: instruction.keys[ataInitInstructionKeysIndexes.PayerAddress].pubkey.toString(),
                        tokenName: mintTokenName,
                        programId: programID,
                    },
                };
                instructionData.push(ataInit);
                break;
            case constants_1.ValidInstructionTypesEnum.CloseAssociatedTokenAccount:
                const accountAddress = instruction.keys[closeAtaInstructionKeysIndexes.AccountAddress].pubkey.toString();
                const destinationAddress = instruction.keys[closeAtaInstructionKeysIndexes.DestinationAddress].pubkey.toString();
                const authorityAddress = instruction.keys[closeAtaInstructionKeysIndexes.AuthorityAddress].pubkey.toString();
                const ataClose = {
                    type: constants_1.InstructionBuilderTypes.CloseAssociatedTokenAccount,
                    params: {
                        accountAddress,
                        destinationAddress,
                        authorityAddress,
                    },
                };
                instructionData.push(ataClose);
                break;
            case constants_1.ValidInstructionTypesEnum.SetComputeUnitLimit:
                const setComputeUnitLimitParams = web3_js_1.ComputeBudgetInstruction.decodeSetComputeUnitLimit(instruction);
                const setComputeUnitLimit = {
                    type: constants_1.InstructionBuilderTypes.SetComputeUnitLimit,
                    params: {
                        units: setComputeUnitLimitParams.units,
                    },
                };
                instructionData.push(setComputeUnitLimit);
                break;
            case constants_1.ValidInstructionTypesEnum.SetPriorityFee:
                const setComputeUnitPriceParams = web3_js_1.ComputeBudgetInstruction.decodeSetComputeUnitPrice(instruction);
                const setPriorityFee = {
                    type: constants_1.InstructionBuilderTypes.SetPriorityFee,
                    params: {
                        fee: setComputeUnitPriceParams.microLamports,
                    },
                };
                instructionData.push(setPriorityFee);
                break;
            case constants_1.ValidInstructionTypesEnum.MintTo:
                let mintToInstruction;
                if (instruction.programId.toString() !== spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
                    mintToInstruction = (0, spl_token_1.decodeMintToInstruction)(instruction);
                }
                else {
                    mintToInstruction = (0, spl_token_1.decodeMintToInstruction)(instruction, spl_token_1.TOKEN_2022_PROGRAM_ID);
                }
                const mintAddressForMint = mintToInstruction.keys.mint.pubkey.toString();
                const tokenNameForMint = findTokenName(mintAddressForMint, instructionMetadata, _useTokenAddressTokenName);
                let programIDForMint;
                if (instruction.programId) {
                    programIDForMint = instruction.programId.toString();
                }
                const mintTo = {
                    type: constants_1.InstructionBuilderTypes.MintTo,
                    params: {
                        mintAddress: mintAddressForMint,
                        destinationAddress: mintToInstruction.keys.destination.pubkey.toString(),
                        authorityAddress: mintToInstruction.keys.authority.pubkey.toString(),
                        amount: mintToInstruction.data.amount.toString(),
                        tokenName: tokenNameForMint,
                        decimalPlaces: undefined,
                        programId: programIDForMint,
                    },
                };
                instructionData.push(mintTo);
                break;
            case constants_1.ValidInstructionTypesEnum.Burn:
                let burnInstruction;
                if (instruction.programId.toString() !== spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
                    burnInstruction = (0, spl_token_1.decodeBurnInstruction)(instruction);
                }
                else {
                    burnInstruction = (0, spl_token_1.decodeBurnInstruction)(instruction, spl_token_1.TOKEN_2022_PROGRAM_ID);
                }
                const mintAddressForBurn = burnInstruction.keys.mint.pubkey.toString();
                const tokenNameForBurn = findTokenName(mintAddressForBurn, instructionMetadata, _useTokenAddressTokenName);
                let programIDForBurn;
                if (instruction.programId) {
                    programIDForBurn = instruction.programId.toString();
                }
                const burn = {
                    type: constants_1.InstructionBuilderTypes.Burn,
                    params: {
                        mintAddress: mintAddressForBurn,
                        accountAddress: burnInstruction.keys.account.pubkey.toString(),
                        authorityAddress: burnInstruction.keys.owner.pubkey.toString(),
                        amount: burnInstruction.data.amount.toString(),
                        tokenName: tokenNameForBurn,
                        decimalPlaces: undefined,
                        programId: programIDForBurn,
                    },
                };
                instructionData.push(burn);
                break;
            default:
                throw new sdk_core_1.NotSupported('Invalid transaction, instruction type not supported: ' + (0, utils_1.getInstructionType)(instruction));
        }
    }
    return instructionData;
}
function isJitoStakingInstructions(si) {
    return si.depositSol !== undefined;
}
function isMarinadeStakingInstructions(si) {
    return si.create !== undefined && si.initialize !== undefined && si.delegate === undefined;
}
function isNativeStakingInstructions(si) {
    return si.create !== undefined && si.initialize !== undefined && si.delegate !== undefined;
}
function getStakingTypeFromStakingInstructions(si) {
    const isJito = isJitoStakingInstructions(si);
    const isMarinade = isMarinadeStakingInstructions(si);
    const isNative = isNativeStakingInstructions(si);
    (0, assert_1.default)([isJito, isMarinade, isNative].filter((x) => x).length === 1, 'StakingType is ambiguous');
    if (isJito)
        return public_types_1.SolStakingTypeEnum.JITO;
    if (isMarinade)
        return public_types_1.SolStakingTypeEnum.MARINADE;
    if (isNative)
        return public_types_1.SolStakingTypeEnum.NATIVE;
    (0, assert_1.default)(false, 'No StakingType found');
}
/**
 * Parses Solana instructions to create staking tx and delegate tx instructions params
 * Only supports Nonce, StakingActivate and Memo Solana instructions
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for staking activate tx
 */
function parseStakingActivateInstructions(instructions) {
    const instructionData = [];
    const stakingInstructions = {};
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.Memo:
                const memo = { type: constants_1.InstructionBuilderTypes.Memo, params: { memo: instruction.data.toString() } };
                instructionData.push(memo);
                break;
            case constants_1.ValidInstructionTypesEnum.Create:
                stakingInstructions.create = web3_js_1.SystemInstruction.decodeCreateAccount(instruction);
                break;
            case constants_1.ValidInstructionTypesEnum.StakingInitialize:
                stakingInstructions.initialize = web3_js_1.StakeInstruction.decodeInitialize(instruction);
                break;
            case constants_1.ValidInstructionTypesEnum.StakingDelegate:
                stakingInstructions.delegate = web3_js_1.StakeInstruction.decodeDelegate(instruction);
                break;
            case constants_1.ValidInstructionTypesEnum.DepositSol:
                stakingInstructions.depositSol = (0, jitoStakePoolOperations_1.decodeDepositSol)(instruction);
                break;
            case constants_1.ValidInstructionTypesEnum.InitializeAssociatedTokenAccount:
                stakingInstructions.hasAtaInit = true;
                instructionData.push({
                    type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                    params: {
                        mintAddress: instruction.keys[ataInitInstructionKeysIndexes.MintAddress].pubkey.toString(),
                        ataAddress: instruction.keys[ataInitInstructionKeysIndexes.ATAAddress].pubkey.toString(),
                        ownerAddress: instruction.keys[ataInitInstructionKeysIndexes.OwnerAddress].pubkey.toString(),
                        payerAddress: instruction.keys[ataInitInstructionKeysIndexes.PayerAddress].pubkey.toString(),
                        tokenName: findTokenName(instruction.keys[ataInitInstructionKeysIndexes.MintAddress].pubkey.toString()),
                    },
                });
                break;
        }
    }
    validateStakingInstructions(stakingInstructions);
    const stakingType = getStakingTypeFromStakingInstructions(stakingInstructions);
    let stakingActivate;
    switch (stakingType) {
        case public_types_1.SolStakingTypeEnum.JITO: {
            (0, assert_1.default)(isJitoStakingInstructions(stakingInstructions));
            const { depositSol, hasAtaInit } = stakingInstructions;
            stakingActivate = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    stakingType,
                    fromAddress: depositSol.fundingAccount.toString(),
                    stakingAddress: depositSol.stakePool.toString(),
                    amount: depositSol.lamports.toString(),
                    validator: depositSol.stakePool.toString(),
                    extraParams: {
                        stakePoolData: {
                            managerFeeAccount: depositSol.managerFeeAccount.toString(),
                            poolMint: depositSol.poolMint.toString(),
                            reserveStake: depositSol.reserveStake.toString(),
                        },
                        createAssociatedTokenAccount: !!hasAtaInit,
                    },
                },
            };
            break;
        }
        case public_types_1.SolStakingTypeEnum.MARINADE: {
            (0, assert_1.default)(isMarinadeStakingInstructions(stakingInstructions));
            const { create, initialize } = stakingInstructions;
            stakingActivate = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    stakingType,
                    fromAddress: create.fromPubkey.toString(),
                    stakingAddress: initialize.stakePubkey.toString(),
                    amount: create.lamports.toString(),
                    validator: initialize.authorized.staker.toString(),
                },
            };
            break;
        }
        case public_types_1.SolStakingTypeEnum.NATIVE: {
            (0, assert_1.default)(isNativeStakingInstructions(stakingInstructions));
            const { create, initialize, delegate } = stakingInstructions;
            stakingActivate = {
                type: constants_1.InstructionBuilderTypes.StakingActivate,
                params: {
                    stakingType,
                    fromAddress: create.fromPubkey.toString(),
                    stakingAddress: initialize.stakePubkey.toString(),
                    amount: create.lamports.toString(),
                    validator: delegate.votePubkey.toString(),
                },
            };
            break;
        }
        default: {
            const unreachable = stakingType;
            throw new Error(`Unknown staking type ${unreachable}`);
        }
    }
    instructionData.push(stakingActivate);
    return instructionData;
}
/**
 * Parses Solana instructions to create delegate tx
 * Only supports Nonce, StakingDelegate
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for staking delegate tx
 */
function parseStakingDelegateInstructions(instructions) {
    const instructionData = [];
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.StakingDelegate:
                const stakingDelegateParams = web3_js_1.StakeInstruction.decodeDelegate(instruction);
                const stakingDelegate = {
                    type: constants_1.InstructionBuilderTypes.StakingDelegate,
                    params: {
                        fromAddress: stakingDelegateParams.authorizedPubkey.toString() || '',
                        stakingAddress: stakingDelegateParams.stakePubkey.toString() || '',
                        validator: stakingDelegateParams.votePubkey.toString() || '',
                    },
                };
                instructionData.push(stakingDelegate);
                break;
        }
    }
    return instructionData;
}
function validateStakingInstructions(stakingInstructions) {
    if (stakingInstructions.delegate === undefined && stakingInstructions.depositSol !== undefined) {
        return;
    }
    if (!stakingInstructions.create) {
        throw new sdk_core_1.NotSupported('Invalid staking activate transaction, missing create stake account instruction');
    }
    if (!stakingInstructions.delegate && !stakingInstructions.initialize) {
        throw new sdk_core_1.NotSupported('Invalid staking activate transaction, missing initialize stake account/delegate instruction');
    }
}
function isJitoUnstakingInstructions(ui) {
    return ui.withdrawStake !== undefined && ui.deactivate !== undefined;
}
function isMarinadeUnstakingInstructions(ui) {
    return ui.transfer !== undefined && ui.deactivate === undefined;
}
function isNativeUnstakingInstructions(ui) {
    return ui.withdrawStake === undefined && ui.deactivate !== undefined;
}
function getStakingTypeFromUnstakingInstructions(ui) {
    const isJito = isJitoUnstakingInstructions(ui);
    const isMarinade = isMarinadeUnstakingInstructions(ui);
    const isNative = isNativeUnstakingInstructions(ui);
    (0, assert_1.default)([isJito, isMarinade, isNative].filter((x) => x).length === 1, 'StakingType is ambiguous');
    if (isJito)
        return public_types_1.SolStakingTypeEnum.JITO;
    if (isMarinade)
        return public_types_1.SolStakingTypeEnum.MARINADE;
    if (isNative)
        return public_types_1.SolStakingTypeEnum.NATIVE;
    (0, assert_1.default)(false, 'No StakingType found');
}
/**
 * Parses Solana instructions to create deactivate stake tx instructions params. Supports full stake
 * account deactivation and partial stake account deactivation.
 *
 * When partially deactivating a stake account this method expects the following instructions: Allocate,
 * to allocate a new staking account, Assign, to assign the newly created staking account to the
 * Stake Program, Split, to split the current stake account, and StakingDeactivate to deactivate the
 * newly created stake account.
 *
 * Supports Nonce, StakingDeactivate, Memo, Allocate, Assign, and Split Solana instructions.
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for staking deactivate tx
 */
function parseStakingDeactivateInstructions(instructions, coinName) {
    const instructionData = [];
    const unstakingInstructions = [];
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.Memo:
                const memo = {
                    type: constants_1.InstructionBuilderTypes.Memo,
                    params: { memo: instruction.data.toString() },
                };
                instructionData.push(memo);
                break;
            case constants_1.ValidInstructionTypesEnum.Allocate:
                if (unstakingInstructions.length > 0 &&
                    unstakingInstructions[unstakingInstructions.length - 1].allocate === undefined) {
                    unstakingInstructions[unstakingInstructions.length - 1].allocate =
                        web3_js_1.SystemInstruction.decodeAllocate(instruction);
                }
                else {
                    unstakingInstructions.push({
                        allocate: web3_js_1.SystemInstruction.decodeAllocate(instruction),
                    });
                }
                break;
            case constants_1.ValidInstructionTypesEnum.Assign:
                if (unstakingInstructions.length > 0 &&
                    unstakingInstructions[unstakingInstructions.length - 1].assign === undefined) {
                    unstakingInstructions[unstakingInstructions.length - 1].assign = web3_js_1.SystemInstruction.decodeAssign(instruction);
                }
                else {
                    unstakingInstructions.push({
                        assign: web3_js_1.SystemInstruction.decodeAssign(instruction),
                    });
                }
                break;
            case constants_1.ValidInstructionTypesEnum.Split:
                if (unstakingInstructions.length > 0 &&
                    unstakingInstructions[unstakingInstructions.length - 1].split === undefined) {
                    unstakingInstructions[unstakingInstructions.length - 1].split = web3_js_1.StakeInstruction.decodeSplit(instruction);
                }
                else {
                    unstakingInstructions.push({
                        split: web3_js_1.StakeInstruction.decodeSplit(instruction),
                    });
                }
                break;
            case constants_1.ValidInstructionTypesEnum.StakingDeactivate:
                if (unstakingInstructions.length > 0 &&
                    unstakingInstructions[unstakingInstructions.length - 1].deactivate === undefined) {
                    unstakingInstructions[unstakingInstructions.length - 1].deactivate =
                        web3_js_1.StakeInstruction.decodeDeactivate(instruction);
                }
                else {
                    unstakingInstructions.push({
                        deactivate: web3_js_1.StakeInstruction.decodeDeactivate(instruction),
                    });
                }
                break;
            case constants_1.ValidInstructionTypesEnum.Transfer:
                if (unstakingInstructions.length > 0 &&
                    unstakingInstructions[unstakingInstructions.length - 1].transfer === undefined) {
                    unstakingInstructions[unstakingInstructions.length - 1].transfer =
                        web3_js_1.SystemInstruction.decodeTransfer(instruction);
                }
                else {
                    unstakingInstructions.push({
                        transfer: web3_js_1.SystemInstruction.decodeTransfer(instruction),
                    });
                }
                break;
            case constants_1.ValidInstructionTypesEnum.WithdrawStake:
                if (unstakingInstructions.length > 0 &&
                    unstakingInstructions[unstakingInstructions.length - 1].withdrawStake === undefined) {
                    unstakingInstructions[unstakingInstructions.length - 1].withdrawStake = (0, jitoStakePoolOperations_1.decodeWithdrawStake)(instruction);
                }
                else {
                    unstakingInstructions.push({
                        withdrawStake: (0, jitoStakePoolOperations_1.decodeWithdrawStake)(instruction),
                    });
                }
                break;
        }
    }
    for (const unstakingInstruction of unstakingInstructions) {
        validateUnstakingInstructions(unstakingInstruction);
        const stakingType = getStakingTypeFromUnstakingInstructions(unstakingInstruction);
        let stakingDeactivate;
        switch (stakingType) {
            case public_types_1.SolStakingTypeEnum.JITO: {
                (0, assert_1.default)(isJitoUnstakingInstructions(unstakingInstruction));
                const { withdrawStake } = unstakingInstruction;
                stakingDeactivate = {
                    type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                    params: {
                        stakingType,
                        fromAddress: withdrawStake.destinationStakeAuthority.toString(),
                        stakingAddress: withdrawStake.stakePool.toString(),
                        amount: withdrawStake.poolTokens.toString(),
                        unstakingAddress: withdrawStake.destinationStake.toString(),
                        extraParams: {
                            stakePoolData: {
                                managerFeeAccount: withdrawStake.managerFeeAccount.toString(),
                                poolMint: withdrawStake.poolMint.toString(),
                                validatorListAccount: withdrawStake.validatorList.toString(),
                            },
                            validatorAddress: withdrawStake.validatorStake.toString(),
                            transferAuthorityAddress: withdrawStake.sourceTransferAuthority.toString(),
                        },
                    },
                };
                break;
            }
            case public_types_1.SolStakingTypeEnum.MARINADE: {
                (0, assert_1.default)(isMarinadeUnstakingInstructions(unstakingInstruction));
                const { transfer } = unstakingInstruction;
                stakingDeactivate = {
                    type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                    params: {
                        stakingType,
                        fromAddress: '',
                        stakingAddress: '',
                        recipients: [
                            {
                                address: transfer.toPubkey.toString() || '',
                                amount: transfer.lamports.toString() || '',
                            },
                        ],
                    },
                };
                break;
            }
            case public_types_1.SolStakingTypeEnum.NATIVE: {
                (0, assert_1.default)(isNativeUnstakingInstructions(unstakingInstruction));
                const { deactivate, split } = unstakingInstruction;
                stakingDeactivate = {
                    type: constants_1.InstructionBuilderTypes.StakingDeactivate,
                    params: {
                        stakingType,
                        fromAddress: deactivate.authorizedPubkey.toString() || '',
                        stakingAddress: split?.stakePubkey.toString() || deactivate.stakePubkey.toString(),
                        amount: split?.lamports.toString(),
                        unstakingAddress: split?.splitStakePubkey.toString(),
                    },
                };
                break;
            }
            default: {
                const unreachable = stakingType;
                throw new Error(`Unknown staking type ${unreachable}`);
            }
        }
        instructionData.push(stakingDeactivate);
    }
    return instructionData;
}
function validateUnstakingInstructions(unstakingInstructions) {
    // Cases where exactly one field should be present
    const unstakingInstructionsKeys = [
        'allocate',
        'assign',
        'split',
        'deactivate',
        'transfer',
    ];
    if (unstakingInstructionsKeys.every((k) => !!unstakingInstructions[k] === (k === 'transfer'))) {
        return;
    }
    if (unstakingInstructionsKeys.every((k) => !!unstakingInstructions[k] === (k === 'withdrawStake'))) {
        return;
    }
    if (unstakingInstructionsKeys.every((k) => !!unstakingInstructions[k] === (k === 'deactivate'))) {
        return;
    }
    // Cases where deactivate field must be present with another field
    if (!unstakingInstructions.deactivate) {
        throw new sdk_core_1.NotSupported('Invalid deactivate stake transaction, missing deactivate stake account instruction');
    }
    // This is a stake pool instruction, not a partial unstake
    if (unstakingInstructions.withdrawStake) {
        return;
    }
    if (!unstakingInstructions.allocate) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, missing allocate unstake account instruction');
    }
    else if (!unstakingInstructions.assign) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, missing assign unstake account instruction');
    }
    else if (!unstakingInstructions.split) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, missing split stake account instruction');
    }
    else if (unstakingInstructions.allocate.accountPubkey.toString() !== unstakingInstructions.assign.accountPubkey.toString()) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, must allocate and assign the same public key');
    }
    else if (unstakingInstructions.allocate.space !== web3_js_1.StakeProgram.space) {
        throw new sdk_core_1.NotSupported(`Invalid partial deactivate stake transaction, unstaking account must allocate ${web3_js_1.StakeProgram.space} bytes`);
    }
    else if (unstakingInstructions.assign.programId.toString() !== web3_js_1.StakeProgram.programId.toString()) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, the unstake account must be assigned to the Stake Program');
    }
    else if (unstakingInstructions.allocate.accountPubkey.toString() !== unstakingInstructions.split.splitStakePubkey.toString()) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, must allocate the unstaking account');
    }
    else if (unstakingInstructions.split.stakePubkey.toString() === unstakingInstructions.split.splitStakePubkey.toString()) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, the unstaking account must be different from the Stake Account');
    }
    else if (!unstakingInstructions.transfer) {
        throw new sdk_core_1.NotSupported('Invalid partial deactivate stake transaction, missing funding of unstake address instruction');
    }
}
/**
 * Parses Solana instructions to create staking  withdraw tx instructions params
 * Only supports Nonce, StakingWithdraw, and Memo Solana instructions
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for staking withdraw tx
 */
function parseStakingWithdrawInstructions(instructions) {
    const instructionData = [];
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.Memo:
                const memo = {
                    type: constants_1.InstructionBuilderTypes.Memo,
                    params: { memo: instruction.data.toString() },
                };
                instructionData.push(memo);
                break;
            case constants_1.ValidInstructionTypesEnum.StakingWithdraw:
                const withdrawInstruction = web3_js_1.StakeInstruction.decodeWithdraw(instruction);
                const stakingWithdraw = {
                    type: constants_1.InstructionBuilderTypes.StakingWithdraw,
                    params: {
                        fromAddress: withdrawInstruction.authorizedPubkey.toString(),
                        stakingAddress: withdrawInstruction.stakePubkey.toString(),
                        amount: withdrawInstruction.lamports.toString(),
                    },
                };
                instructionData.push(stakingWithdraw);
                break;
        }
    }
    return instructionData;
}
/**
 * Get the memo object from instructions if it exists
 *
 * @param {TransactionInstruction[]} instructions - the array of supported Solana instructions to be parsed
 * @param {Record<string, number>} instructionIndexes - the instructions indexes of the current transaction
 * @returns {Memo | undefined} - memo object or undefined
 */
function getMemo(instructions, instructionIndexes) {
    const instructionsLength = Object.keys(instructionIndexes).length;
    if (instructions.length === instructionsLength && instructions[instructionIndexes.Memo]) {
        return {
            type: constants_1.InstructionBuilderTypes.Memo,
            params: { memo: instructions[instructionIndexes.Memo].data.toString() },
        };
    }
}
const ataInitInstructionKeysIndexes = {
    PayerAddress: 0,
    ATAAddress: 1,
    OwnerAddress: 2,
    MintAddress: 3,
};
const closeAtaInstructionKeysIndexes = {
    AccountAddress: 0,
    DestinationAddress: 1,
    AuthorityAddress: 2,
};
/**
 * Parses Solana instructions to initialize associated token account tx instructions params
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for Send tx
 */
function parseAtaInitInstructions(instructions, instructionMetadata, _useTokenAddressTokenName) {
    const instructionData = [];
    let memo;
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.Memo:
                memo = { type: constants_1.InstructionBuilderTypes.Memo, params: { memo: instruction.data.toString() } };
                break;
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.InitializeAssociatedTokenAccount:
                const mintAddress = instruction.keys[ataInitInstructionKeysIndexes.MintAddress].pubkey.toString();
                const tokenName = findTokenName(mintAddress, instructionMetadata, _useTokenAddressTokenName);
                let programID;
                if (instruction.programId) {
                    programID = instruction.programId.toString();
                }
                const ataInit = {
                    type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                    params: {
                        mintAddress,
                        ataAddress: instruction.keys[ataInitInstructionKeysIndexes.ATAAddress].pubkey.toString(),
                        ownerAddress: instruction.keys[ataInitInstructionKeysIndexes.OwnerAddress].pubkey.toString(),
                        payerAddress: instruction.keys[ataInitInstructionKeysIndexes.PayerAddress].pubkey.toString(),
                        tokenName,
                        programId: programID,
                    },
                };
                instructionData.push(ataInit);
                break;
            case constants_1.ValidInstructionTypesEnum.DepositSol:
                // AtaInit is a part of spl-stake-pool's depositSol process
                break;
            default:
                throw new sdk_core_1.NotSupported('Invalid transaction, instruction type not supported: ' + (0, utils_1.getInstructionType)(instruction));
        }
    }
    if (memo) {
        instructionData.push(memo);
    }
    return instructionData;
}
const ataCloseInstructionKeysIndexes = {
    AccountAddress: 0,
    DestinationAddress: 1,
    AuthorityAddress: 2,
};
/**
 * Parses Solana instructions to close associated token account tx instructions params
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for Send tx
 */
function parseAtaCloseInstructions(instructions) {
    const instructionData = [];
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.CloseAssociatedTokenAccount:
                const ataClose = {
                    type: constants_1.InstructionBuilderTypes.CloseAssociatedTokenAccount,
                    params: {
                        accountAddress: instruction.keys[ataCloseInstructionKeysIndexes.AccountAddress].pubkey.toString(),
                        destinationAddress: instruction.keys[ataCloseInstructionKeysIndexes.DestinationAddress].pubkey.toString(),
                        authorityAddress: instruction.keys[ataCloseInstructionKeysIndexes.AuthorityAddress].pubkey.toString(),
                    },
                };
                instructionData.push(ataClose);
                break;
            default:
                throw new sdk_core_1.NotSupported('Invalid transaction, instruction type not supported: ' + (0, utils_1.getInstructionType)(instruction));
        }
    }
    return instructionData;
}
/**
 * Parses Solana instructions to authorized staking account params
 * Only supports Nonce, Authorize instructions
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for staking authorize tx
 */
function parseStakingAuthorizeInstructions(instructions) {
    const instructionData = [];
    for (const instruction of instructions) {
        const type = (0, utils_1.getInstructionType)(instruction);
        switch (type) {
            case constants_1.ValidInstructionTypesEnum.AdvanceNonceAccount:
                const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instruction);
                const nonce = {
                    type: constants_1.InstructionBuilderTypes.NonceAdvance,
                    params: {
                        walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
                        authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
                    },
                };
                instructionData.push(nonce);
                break;
            case constants_1.ValidInstructionTypesEnum.Memo:
                const memo = { type: constants_1.InstructionBuilderTypes.Memo, params: { memo: instruction.data.toString() } };
                instructionData.push(memo);
                break;
            case constants_1.ValidInstructionTypesEnum.Authorize:
                const authorize = web3_js_1.StakeInstruction.decodeAuthorize(instruction);
                instructionData.push({
                    type: constants_1.InstructionBuilderTypes.StakingAuthorize,
                    params: {
                        stakingAddress: authorize.stakePubkey.toString(),
                        oldAuthorizeAddress: authorize.authorizedPubkey.toString(),
                        newAuthorizeAddress: authorize.newAuthorizedPubkey.toString(),
                        newWithdrawAddress: authorize.custodianPubkey?.toString() || '',
                    },
                });
                break;
        }
    }
    return instructionData;
}
/**
 * Parses Solana instructions to authorized staking account params
 * Only supports Nonce, Authorize instructions
 *
 * @param {TransactionInstruction[]} instructions - an array of supported Solana instructions
 * @returns {InstructionParams[]} An array containing instruction params for staking authorize tx
 */
function parseStakingAuthorizeRawInstructions(instructions) {
    const instructionData = [];
    (0, assert_1.default)(instructions.length === 2, 'Invalid number of instructions');
    const advanceNonceInstruction = web3_js_1.SystemInstruction.decodeNonceAdvance(instructions[0]);
    const nonce = {
        type: constants_1.InstructionBuilderTypes.NonceAdvance,
        params: {
            walletNonceAddress: advanceNonceInstruction.noncePubkey.toString(),
            authWalletAddress: advanceNonceInstruction.authorizedPubkey.toString(),
        },
    };
    instructionData.push(nonce);
    const authorize = instructions[1];
    (0, assert_1.default)(authorize.keys.length === 5, 'Invalid number of keys in authorize instruction');
    instructionData.push({
        type: constants_1.InstructionBuilderTypes.StakingAuthorize,
        params: {
            stakingAddress: authorize.keys[0].pubkey.toString(),
            oldAuthorizeAddress: authorize.keys[2].pubkey.toString(),
            newAuthorizeAddress: authorize.keys[3].pubkey.toString(),
            custodianAddress: authorize.keys[4].pubkey.toString(),
        },
    });
    return instructionData;
}
/**
 * Parses Solana instructions to custom instruction params
 *
 * @param {TransactionInstruction[]} instructions - containing custom solana instructions
 * @param {InstructionParams[]} instructionMetadata - the instruction metadata for the transaction
 * @returns {InstructionParams[]} An array containing instruction params for custom instructions
 */
function parseCustomInstructions(instructions, instructionMetadata) {
    const instructionData = [];
    for (let i = 0; i < instructions.length; i++) {
        const instruction = instructions[i];
        // Check if we have metadata for this instruction position
        if (instructionMetadata &&
            instructionMetadata[i] &&
            instructionMetadata[i].type === constants_1.InstructionBuilderTypes.CustomInstruction) {
            instructionData.push(instructionMetadata[i]);
        }
        else {
            // Convert the raw instruction to CustomInstruction format
            const customInstruction = {
                type: constants_1.InstructionBuilderTypes.CustomInstruction,
                params: {
                    programId: instruction.programId.toString(),
                    keys: instruction.keys.map((key) => ({
                        pubkey: key.pubkey.toString(),
                        isSigner: key.isSigner,
                        isWritable: key.isWritable,
                    })),
                    data: instruction.data.toString('base64'),
                },
            };
            instructionData.push(customInstruction);
        }
    }
    return instructionData;
}
function findTokenName(mintAddress, instructionMetadata, _useTokenAddressTokenName) {
    let token;
    statics_1.coins.forEach((value, key) => {
        if (value instanceof statics_1.SolCoin && value.tokenAddress === mintAddress) {
            token = value.name;
        }
    });
    if (!token && instructionMetadata) {
        instructionMetadata.forEach((instruction) => {
            if (instruction.type === constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount &&
                instruction.params.mintAddress === mintAddress) {
                token = instruction.params.tokenName;
            }
            else if (instruction.type === constants_1.InstructionBuilderTypes.TokenTransfer &&
                instruction.params.tokenAddress === mintAddress) {
                token = instruction.params.tokenName;
            }
        });
    }
    if (!token && _useTokenAddressTokenName) {
        token = mintAddress;
    }
    (0, assert_1.default)(token);
    return token;
}
