// @bitgo/sdk-coin-sol/dist/cjs/src/lib/jitoStakePoolOperations.js
"use strict";
/**
 * @file Implementation of the depositSol instruction. On upgrade of
 * '@solana/spl-token', this module may no longer be necessary.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WITHDRAW_STAKE_LAYOUT_CODE = exports.DEPOSIT_SOL_LAYOUT_CODE = void 0;
exports.findWithdrawAuthorityProgramAddressSync = findWithdrawAuthorityProgramAddressSync;
exports.depositSolInstructions = depositSolInstructions;
exports.decodeDepositSol = decodeDepositSol;
exports.withdrawStakeInstructions = withdrawStakeInstructions;
exports.decodeWithdrawStake = decodeWithdrawStake;
const spl_stake_pool_1 = require("@solana/spl-stake-pool");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const assert_1 = __importDefault(require("assert"));
const constants_1 = require("./constants");
exports.DEPOSIT_SOL_LAYOUT_CODE = 14;
exports.WITHDRAW_STAKE_LAYOUT_CODE = 10;
/**
 * Generates the withdraw authority program address for the stake pool.
 * Like findWithdrawAuthorityProgramAddress in @solana/spl-stake-pool,
 * but synchronous.
 *
 * @see {findWithdrawAuthorityProgramAddress}
 */
function findWithdrawAuthorityProgramAddressSync(programId, stakePoolAddress) {
    const [withdrawAuthority] = web3_js_1.PublicKey.findProgramAddressSync([stakePoolAddress.toBuffer(), Buffer.from('withdraw')], programId);
    return withdrawAuthority;
}
/**
 * Construct Solana depositSol stake pool instruction from parameters.
 *
 * @param {DepositSolInstructionsParams} params - parameters for staking to stake pool
 * @param {DepositSolStakePoolData} stakePool - data from getStakePoolAccount needed for DepositSol
 * @param createAssociatedTokenAccount
 * @returns {TransactionInstruction}
 */
function depositSolInstructions(params, stakePool, createAssociatedTokenAccount) {
    const { stakePoolAddress, from, lamports } = params;
    const poolMint = new web3_js_1.PublicKey(stakePool.poolMint);
    const reserveStake = new web3_js_1.PublicKey(stakePool.reserveStake);
    const managerFeeAccount = new web3_js_1.PublicKey(stakePool.managerFeeAccount);
    // findWithdrawAuthorityProgramAddress
    const withdrawAuthority = findWithdrawAuthorityProgramAddressSync(spl_stake_pool_1.STAKE_POOL_PROGRAM_ID, stakePoolAddress);
    const associatedAddress = (0, spl_token_1.getAssociatedTokenAddressSync)(poolMint, from);
    const instructions = [];
    if (createAssociatedTokenAccount) {
        instructions.push((0, spl_token_1.createAssociatedTokenAccountInstruction)(from, associatedAddress, from, poolMint));
    }
    instructions.push(spl_stake_pool_1.StakePoolInstruction.depositSol({
        stakePool: stakePoolAddress,
        reserveStake,
        fundingAccount: from,
        destinationPoolAccount: associatedAddress,
        managerFeeAccount,
        referralPoolAccount: associatedAddress,
        poolMint,
        lamports: Number(lamports),
        withdrawAuthority,
    }));
    return instructions;
}
function parseKey(key, template) {
    (0, assert_1.default)(key.isSigner === template.isSigner && key.isWritable === template.isWritable, `Unexpected key metadata in instruction: { isSigner: ${key.isSigner}, isWritable: ${key.isWritable} }`);
    return key.pubkey;
}
/**
 * Construct Solana depositSol stake pool parameters from instruction.
 *
 * @param {TransactionInstruction} instruction
 * @returns {DepositSolParams}
 */
function decodeDepositSol(instruction) {
    const { programId, keys, data } = instruction;
    (0, assert_1.default)(programId.equals(spl_stake_pool_1.STAKE_POOL_PROGRAM_ID), 'Invalid DepositSol instruction, program ID must be the Stake Pool Program');
    let i = 0;
    const stakePool = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const withdrawAuthority = parseKey(keys[i++], { isSigner: false, isWritable: false });
    const reserveStake = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const fundingAccount = parseKey(keys[i++], { isSigner: true, isWritable: true });
    const destinationPoolAccount = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const managerFeeAccount = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const referralPoolAccount = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const poolMint = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const systemProgramProgramId = parseKey(keys[i++], { isSigner: false, isWritable: false });
    (0, assert_1.default)(systemProgramProgramId.equals(web3_js_1.SystemProgram.programId), 'Unexpected pubkey in DepositSol instruction');
    const tokenProgramId = parseKey(keys[i++], { isSigner: false, isWritable: false });
    (0, assert_1.default)(tokenProgramId.equals(spl_token_1.TOKEN_PROGRAM_ID), 'Unexpected pubkey in DepositSol instruction');
    const depositAuthority = keys.length > 10 ? parseKey(keys[i++], { isSigner: true, isWritable: false }) : undefined;
    (0, assert_1.default)(keys.length <= 11, 'Too many keys in DepositSol instruction');
    const layoutCode = data.readUint8(0);
    (0, assert_1.default)(layoutCode === exports.DEPOSIT_SOL_LAYOUT_CODE, 'Incorrect layout code in DepositSol data');
    (0, assert_1.default)(data.length === 9, 'Incorrect data size for DepositSol layout');
    const lamports = data.readBigInt64LE(1);
    return {
        stakePool,
        depositAuthority,
        withdrawAuthority,
        reserveStake,
        fundingAccount,
        destinationPoolAccount,
        managerFeeAccount,
        referralPoolAccount,
        poolMint,
        lamports: Number(lamports),
    };
}
/**
 * Construct Solana withdrawStake stake pool instructions from parameters.
 *
 * @param {WithdrawStakeInstructionsParams} params - parameters for unstaking from stake pool
 * @param {WithdrawStakeStakePoolData} stakePool - data from getStakePoolAccount needed for WithdrawStake
 * @returns {TransactionInstruction[]}
 */
function withdrawStakeInstructions(params, stakePool) {
    const { tokenOwner, stakePoolAddress, destinationStakeAccount, validatorAddress, transferAuthority, poolAmount: poolAmountString, } = params;
    const poolMint = new web3_js_1.PublicKey(stakePool.poolMint);
    const validatorList = new web3_js_1.PublicKey(stakePool.validatorListAccount);
    const managerFeeAccount = new web3_js_1.PublicKey(stakePool.managerFeeAccount);
    const poolTokenAccount = (0, spl_token_1.getAssociatedTokenAddressSync)(poolMint, tokenOwner);
    const withdrawAuthority = findWithdrawAuthorityProgramAddressSync(spl_stake_pool_1.STAKE_POOL_PROGRAM_ID, stakePoolAddress);
    const poolAmount = BigInt(poolAmountString);
    return [
        (0, spl_token_1.createApproveInstruction)(poolTokenAccount, transferAuthority, tokenOwner, poolAmount),
        web3_js_1.SystemProgram.createAccount({
            fromPubkey: tokenOwner,
            newAccountPubkey: destinationStakeAccount,
            lamports: constants_1.STAKE_ACCOUNT_RENT_EXEMPT_AMOUNT,
            space: web3_js_1.StakeProgram.space,
            programId: web3_js_1.StakeProgram.programId,
        }),
        spl_stake_pool_1.StakePoolInstruction.withdrawStake({
            stakePool: stakePoolAddress,
            validatorList: validatorList,
            validatorStake: validatorAddress,
            destinationStake: destinationStakeAccount,
            destinationStakeAuthority: tokenOwner,
            sourceTransferAuthority: transferAuthority,
            sourcePoolAccount: poolTokenAccount,
            managerFeeAccount: managerFeeAccount,
            poolMint: poolMint,
            poolTokens: Number(poolAmount),
            withdrawAuthority,
        }),
        ...web3_js_1.StakeProgram.deactivate({
            stakePubkey: destinationStakeAccount,
            authorizedPubkey: tokenOwner,
        }).instructions,
    ];
}
/**
 * Construct Solana withdrawStake stake pool parameters from instruction.
 *
 * @param {TransactionInstruction} instruction
 * @returns {DepositSolParams}
 */
function decodeWithdrawStake(instruction) {
    const { programId, keys, data } = instruction;
    (0, assert_1.default)(programId.equals(spl_stake_pool_1.STAKE_POOL_PROGRAM_ID), 'Invalid WithdrawStake instruction, program ID must be the Stake Pool Program');
    const layoutCode = data.readUint8(0);
    (0, assert_1.default)(layoutCode === exports.WITHDRAW_STAKE_LAYOUT_CODE, 'Incorrect layout code in WithdrawStake data');
    (0, assert_1.default)(data.length === 9, 'Incorrect data size for WithdrawStake layout');
    const poolTokens = data.readBigInt64LE(1);
    let i = 0;
    const stakePool = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const validatorList = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const withdrawAuthority = parseKey(keys[i++], { isSigner: false, isWritable: false });
    const validatorStake = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const destinationStake = keys[i++].pubkey;
    const destinationStakeAuthority = keys[i++].pubkey;
    const sourceTransferAuthority = parseKey(keys[i++], { isSigner: true, isWritable: false });
    const sourcePoolAccount = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const managerFeeAccount = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const poolMint = parseKey(keys[i++], { isSigner: false, isWritable: true });
    const sysvarClockPubkey = parseKey(keys[i++], { isSigner: false, isWritable: false });
    (0, assert_1.default)(sysvarClockPubkey.equals(web3_js_1.SYSVAR_CLOCK_PUBKEY), 'Unexpected pubkey in WithdrawStake instruction');
    const tokenProgramId = parseKey(keys[i++], { isSigner: false, isWritable: false });
    (0, assert_1.default)(tokenProgramId.equals(spl_token_1.TOKEN_PROGRAM_ID), 'Unexpected pubkey in WithdrawStake instruction');
    const stakeProgramProgramId = parseKey(keys[i++], { isSigner: false, isWritable: false });
    (0, assert_1.default)(stakeProgramProgramId.equals(web3_js_1.StakeProgram.programId), 'Unexpected pubkey in WithdrawStake instruction');
    (0, assert_1.default)(i === keys.length, 'Too many keys in WithdrawStake instruction');
    return {
        stakePool,
        validatorList,
        withdrawAuthority,
        validatorStake,
        destinationStake,
        destinationStakeAuthority,
        sourceTransferAuthority,
        sourcePoolAccount,
        managerFeeAccount,
        poolMint,
        poolTokens: Number(poolTokens),
    };
}
