// @bitgo/sdk-coin-sol/dist/cjs/src/lib/solInstructionFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.solInstructionFactory = solInstructionFactory;
const public_types_1 = require("@bitgo/public-types");
const statics_1 = require("@bitgo/statics");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const assert_1 = __importDefault(require("assert"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const constants_1 = require("./constants");
const utils_1 = require("./utils");
const jitoStakePoolOperations_1 = require("./jitoStakePoolOperations");
const token2022Config_1 = require("./token2022Config");
/**
 * Construct Solana instructions from instructions params
 *
 * @param {InstructionParams} instructionToBuild - the data containing the instruction params
 * @returns {TransactionInstruction[]} An array containing supported Solana instructions
 */
function solInstructionFactory(instructionToBuild) {
    switch (instructionToBuild.type) {
        case constants_1.InstructionBuilderTypes.NonceAdvance:
            return advanceNonceInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.Memo:
            return memoInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.Transfer:
            return transferInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.TokenTransfer:
            return tokenTransferInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.Approve:
            return approveInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.CreateNonceAccount:
            return createNonceAccountInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingActivate:
            return stakingInitializeInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingDeactivate:
            return stakingDeactivateInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingWithdraw:
            return stakingWithdrawInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount:
            return createATAInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.CloseAssociatedTokenAccount:
            return closeATAInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingAuthorize:
            return stakingAuthorizeInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingDelegate:
            return stakingDelegateInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.SetComputeUnitLimit:
            return setComputeUnitLimitInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.SetPriorityFee:
            return fetchPriorityFeeInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.MintTo:
            return mintToInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.Burn:
            return burnInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.CustomInstruction:
            return customInstruction(instructionToBuild);
        default:
            throw new Error(`Invalid instruction type or not supported`);
    }
}
/**
 * Construct Advance Nonce Solana instructions
 *
 * @param {Nonce} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Advance Nonce Solana instruction
 */
function advanceNonceInstruction(data) {
    const { params: { authWalletAddress, walletNonceAddress }, } = data;
    (0, assert_1.default)(authWalletAddress, 'Missing authWalletAddress param');
    (0, assert_1.default)(walletNonceAddress, 'Missing walletNonceAddress param');
    const nonceInstruction = web3_js_1.SystemProgram.nonceAdvance({
        noncePubkey: new web3_js_1.PublicKey(walletNonceAddress),
        authorizedPubkey: new web3_js_1.PublicKey(authWalletAddress),
    });
    return [nonceInstruction];
}
function setComputeUnitLimitInstruction(instructionToBuild) {
    const setComputeUnitLimit = web3_js_1.ComputeBudgetProgram.setComputeUnitLimit({
        units: instructionToBuild.params.units,
    });
    return [setComputeUnitLimit];
}
function fetchPriorityFeeInstruction(instructionToBuild) {
    const addPriorityFee = web3_js_1.ComputeBudgetProgram.setComputeUnitPrice({
        microLamports: instructionToBuild.params.fee,
    });
    return [addPriorityFee];
}
/**
 * Construct Memo Solana instructions
 *
 * @param {Memo} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Memo Solana instruction
 */
function memoInstruction(data) {
    const { params: { memo }, } = data;
    (0, assert_1.default)(memo, 'Missing memo param');
    const memoInstruction = new web3_js_1.TransactionInstruction({
        keys: [],
        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
        data: Buffer.from(memo),
    });
    return [memoInstruction];
}
/**
 * Construct Transfer Solana instructions
 *
 * @param {Transfer} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Transfer Solana instruction
 */
function transferInstruction(data) {
    const { params: { fromAddress, toAddress, amount }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress param');
    (0, assert_1.default)(toAddress, 'Missing toAddress param');
    (0, assert_1.default)(amount, 'Missing toAddress param');
    const transferInstruction = web3_js_1.SystemProgram.transfer({
        fromPubkey: new web3_js_1.PublicKey(fromAddress),
        toPubkey: new web3_js_1.PublicKey(toAddress),
        lamports: parseInt(amount, 10),
    });
    return [transferInstruction];
}
/**
 * Construct Transfer Solana instructions
 *
 * @param {Transfer} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Transfer Solana instruction
 */
function tokenTransferInstruction(data) {
    const { params: { fromAddress, toAddress, amount, tokenName, sourceAddress }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress (owner) param');
    (0, assert_1.default)(toAddress, 'Missing toAddress param');
    (0, assert_1.default)(amount, 'Missing amount param');
    (0, assert_1.default)(tokenName, 'Missing token name');
    (0, assert_1.default)(sourceAddress, 'Missing ata address');
    const token = (0, utils_1.getSolTokenFromTokenName)(data.params.tokenName);
    let tokenAddress;
    let programId;
    let decimalPlaces;
    if (data.params.tokenAddress && data.params.decimalPlaces) {
        tokenAddress = data.params.tokenAddress;
        decimalPlaces = data.params.decimalPlaces;
        programId = data.params.programId;
    }
    else if (token) {
        (0, assert_1.default)(token instanceof statics_1.SolCoin);
        tokenAddress = token.tokenAddress;
        decimalPlaces = token.decimalPlaces;
        programId = token.programId;
    }
    else {
        throw new Error('Invalid token name, got:' + data.params.tokenName);
    }
    let transferInstruction;
    const instructions = [];
    if (programId === spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
        // Create the base transfer instruction
        transferInstruction = (0, spl_token_1.createTransferCheckedInstruction)(new web3_js_1.PublicKey(sourceAddress), new web3_js_1.PublicKey(tokenAddress), new web3_js_1.PublicKey(toAddress), new web3_js_1.PublicKey(fromAddress), BigInt(amount), decimalPlaces, [], spl_token_1.TOKEN_2022_PROGRAM_ID);
        // Check if this token has a transfer hook configuration
        const tokenConfig = (0, token2022Config_1.getToken2022Config)(tokenAddress);
        if (tokenConfig?.transferHook) {
            addTransferHookAccounts(transferInstruction, tokenConfig.transferHook);
        }
    }
    else {
        transferInstruction = (0, spl_token_1.createTransferCheckedInstruction)(new web3_js_1.PublicKey(sourceAddress), new web3_js_1.PublicKey(tokenAddress), new web3_js_1.PublicKey(toAddress), new web3_js_1.PublicKey(fromAddress), BigInt(amount), decimalPlaces);
    }
    instructions.push(transferInstruction);
    return instructions;
}
/**
 * Construct Transfer Solana instructions
 *
 * @param {Transfer} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Transfer Solana instruction
 */
function approveInstruction(data) {
    const { params: { accountAddress, delegateAddress, ownerAddress, amount, programId }, } = data;
    (0, assert_1.default)(accountAddress, 'Missing fromAddress (owner) param');
    (0, assert_1.default)(delegateAddress, 'Missing toAddress param');
    (0, assert_1.default)(ownerAddress, 'Missing ownerAddress param');
    (0, assert_1.default)(programId, 'Missing programId param');
    (0, assert_1.default)(amount, 'Missing amount param');
    return [
        (0, spl_token_1.createApproveInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(delegateAddress), new web3_js_1.PublicKey(ownerAddress), BigInt(amount), undefined, programId === undefined ? undefined : new web3_js_1.PublicKey(programId)),
    ];
}
/**
 * Construct Create and Initialize Nonce Solana instructions
 *
 * @param {WalletInit} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Create and Initialize Nonce Solana instruction
 */
function createNonceAccountInstruction(data) {
    const { params: { fromAddress, nonceAddress, authAddress, amount }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress param');
    (0, assert_1.default)(nonceAddress, 'Missing nonceAddress param');
    (0, assert_1.default)(authAddress, 'Missing authAddress param');
    (0, assert_1.default)(amount, 'Missing amount param');
    const nonceAccountInstruction = web3_js_1.SystemProgram.createNonceAccount({
        fromPubkey: new web3_js_1.PublicKey(fromAddress),
        noncePubkey: new web3_js_1.PublicKey(nonceAddress),
        authorizedPubkey: new web3_js_1.PublicKey(authAddress),
        lamports: new bignumber_js_1.default(amount).toNumber(),
    });
    return nonceAccountInstruction.instructions;
}
/**
 * Construct Create Staking Account and Delegate Solana instructions
 *
 * @param {StakingActivate} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Create Staking Account and Delegate Solana instructions
 */
function stakingInitializeInstruction(data) {
    const { params: { fromAddress, stakingAddress, amount, validator, stakingType, extraParams }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress param');
    (0, assert_1.default)(stakingAddress, 'Missing stakingAddress param');
    (0, assert_1.default)(amount, 'Missing amount param');
    (0, assert_1.default)(validator, 'Missing validator param');
    const fromPubkey = new web3_js_1.PublicKey(fromAddress);
    const stakePubkey = new web3_js_1.PublicKey(stakingAddress);
    const validatorPubkey = new web3_js_1.PublicKey(validator);
    const tx = new web3_js_1.Transaction();
    switch (stakingType) {
        case public_types_1.SolStakingTypeEnum.JITO: {
            (0, assert_1.default)(extraParams !== undefined, 'Missing extraParams param');
            const instructions = (0, jitoStakePoolOperations_1.depositSolInstructions)({
                stakePoolAddress: stakePubkey,
                from: fromPubkey,
                lamports: BigInt(amount),
            }, extraParams.stakePoolData, !!extraParams.createAssociatedTokenAccount);
            tx.add(...instructions);
            break;
        }
        case public_types_1.SolStakingTypeEnum.MARINADE: {
            const walletInitStaking = web3_js_1.StakeProgram.createAccount({
                fromPubkey,
                stakePubkey,
                authorized: new web3_js_1.Authorized(validatorPubkey, fromPubkey), // staker and withdrawer
                lockup: new web3_js_1.Lockup(0, 0, fromPubkey), // No minimum epoch to withdraw
                lamports: new bignumber_js_1.default(amount).toNumber(),
            });
            tx.add(walletInitStaking);
            break;
        }
        case public_types_1.SolStakingTypeEnum.NATIVE: {
            const walletInitStaking = web3_js_1.StakeProgram.createAccount({
                fromPubkey,
                stakePubkey,
                authorized: new web3_js_1.Authorized(fromPubkey, fromPubkey), // staker and withdrawer
                lockup: new web3_js_1.Lockup(0, 0, fromPubkey), // No minimum epoch to withdraw
                lamports: new bignumber_js_1.default(amount).toNumber(),
            });
            tx.add(walletInitStaking);
            const delegateStaking = web3_js_1.StakeProgram.delegate({
                stakePubkey: new web3_js_1.PublicKey(stakingAddress),
                authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
                votePubkey: new web3_js_1.PublicKey(validator),
            });
            tx.add(delegateStaking);
            break;
        }
        default: {
            const unreachable = stakingType;
            throw new Error(`Unknown staking type ${unreachable}`);
        }
    }
    return tx.instructions;
}
/**
 * Construct staking deactivate Solana instructions
 *
 * @param {StakingDeactivate} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing staking deactivate instruction
 */
function stakingDeactivateInstruction(data) {
    const { params: { fromAddress, stakingAddress, amount, unstakingAddress, recipients, stakingType, extraParams }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress param');
    switch (stakingType) {
        case public_types_1.SolStakingTypeEnum.JITO: {
            (0, assert_1.default)(stakingAddress, 'Missing stakingAddress param');
            (0, assert_1.default)(unstakingAddress, 'Missing unstakingAddress param');
            (0, assert_1.default)(amount, 'Missing amount param');
            (0, assert_1.default)(extraParams, 'Missing extraParams param');
            const tx = new web3_js_1.Transaction();
            tx.add(...(0, jitoStakePoolOperations_1.withdrawStakeInstructions)({
                stakePoolAddress: new web3_js_1.PublicKey(stakingAddress),
                tokenOwner: new web3_js_1.PublicKey(fromAddress),
                destinationStakeAccount: new web3_js_1.PublicKey(unstakingAddress),
                validatorAddress: new web3_js_1.PublicKey(extraParams.validatorAddress),
                transferAuthority: new web3_js_1.PublicKey(extraParams.transferAuthorityAddress),
                poolAmount: amount,
            }, extraParams.stakePoolData));
            return tx.instructions;
        }
        case public_types_1.SolStakingTypeEnum.MARINADE: {
            (0, assert_1.default)(recipients, 'Missing recipients param');
            const tx = new web3_js_1.Transaction();
            const toPubkeyAddress = new web3_js_1.PublicKey(recipients[0].address || '');
            const transferInstruction = web3_js_1.SystemProgram.transfer({
                fromPubkey: new web3_js_1.PublicKey(fromAddress),
                toPubkey: toPubkeyAddress,
                lamports: parseInt(recipients[0].amount, 10),
            });
            tx.add(transferInstruction);
            return tx.instructions;
        }
        case public_types_1.SolStakingTypeEnum.NATIVE: {
            (0, assert_1.default)(stakingAddress, 'Missing stakingAddress param');
            if (data.params.amount && data.params.unstakingAddress) {
                const tx = new web3_js_1.Transaction();
                const unstakingAddress = new web3_js_1.PublicKey(data.params.unstakingAddress);
                const allocateAccount = web3_js_1.SystemProgram.allocate({
                    accountPubkey: unstakingAddress,
                    space: web3_js_1.StakeProgram.space,
                });
                tx.add(allocateAccount);
                const assignAccount = web3_js_1.SystemProgram.assign({
                    accountPubkey: unstakingAddress,
                    programId: web3_js_1.StakeProgram.programId,
                });
                tx.add(assignAccount);
                const splitStake = web3_js_1.StakeProgram.split({
                    stakePubkey: new web3_js_1.PublicKey(stakingAddress),
                    authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
                    splitStakePubkey: unstakingAddress,
                    lamports: new bignumber_js_1.default(data.params.amount).toNumber(),
                }, 0);
                tx.add(splitStake.instructions[1]);
                const deactivateStaking = web3_js_1.StakeProgram.deactivate({
                    stakePubkey: unstakingAddress,
                    authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
                });
                tx.add(deactivateStaking);
                return tx.instructions;
            }
            else {
                const deactivateStaking = web3_js_1.StakeProgram.deactivate({
                    stakePubkey: new web3_js_1.PublicKey(stakingAddress),
                    authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
                });
                return deactivateStaking.instructions;
            }
        }
        default: {
            const unreachable = stakingType;
            throw new Error(`Unknown staking type ${unreachable}`);
        }
    }
}
/**
 * Construct Staking Withdraw Solana instructions
 *
 * @param {StakingWithdraw} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Staking Withdraw  Solana instructions
 */
function stakingWithdrawInstruction(data) {
    const { params: { fromAddress, stakingAddress, amount }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress param');
    (0, assert_1.default)(stakingAddress, 'Missing stakingAddress param');
    (0, assert_1.default)(amount, 'Missing amount param');
    const withdrawStaking = web3_js_1.StakeProgram.withdraw({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
        toPubkey: new web3_js_1.PublicKey(fromAddress),
        lamports: new bignumber_js_1.default(amount).toNumber(),
    });
    return withdrawStaking.instructions;
}
/**
 * Construct Create and Initialize Nonce Solana instructions
 *
 * @param {WalletInit} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Create and Initialize Nonce Solana instruction
 */
function createATAInstruction(data) {
    const { params: { mintAddress, ataAddress, ownerAddress, payerAddress, programId }, } = data;
    (0, assert_1.default)(mintAddress, 'Missing mintAddress param');
    (0, assert_1.default)(ataAddress, 'Missing ataAddress param');
    (0, assert_1.default)(ownerAddress, 'Missing ownerAddress param');
    (0, assert_1.default)(payerAddress, 'Missing payerAddress param');
    let associatedTokenAccountInstruction;
    if (programId && programId === spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
        associatedTokenAccountInstruction = (0, spl_token_1.createAssociatedTokenAccountInstruction)(new web3_js_1.PublicKey(payerAddress), new web3_js_1.PublicKey(ataAddress), new web3_js_1.PublicKey(ownerAddress), new web3_js_1.PublicKey(mintAddress), spl_token_1.TOKEN_2022_PROGRAM_ID);
    }
    else {
        associatedTokenAccountInstruction = (0, spl_token_1.createAssociatedTokenAccountInstruction)(new web3_js_1.PublicKey(payerAddress), new web3_js_1.PublicKey(ataAddress), new web3_js_1.PublicKey(ownerAddress), new web3_js_1.PublicKey(mintAddress));
    }
    return [associatedTokenAccountInstruction];
}
/**
 * Construct Close ATA Solana instructions
 *
 * @param {WalletInit} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Close ATA Solana instruction
 */
function closeATAInstruction(data) {
    const { params: { accountAddress, destinationAddress, authorityAddress }, } = data;
    (0, assert_1.default)(accountAddress, 'Missing accountAddress param');
    (0, assert_1.default)(destinationAddress, 'Missing destinationAddress param');
    (0, assert_1.default)(authorityAddress, 'Missing authorityAddress param');
    const closeAssociatedTokenAccountInstruction = (0, spl_token_1.createCloseAccountInstruction)(new web3_js_1.PublicKey(accountAddress), new web3_js_1.PublicKey(destinationAddress), new web3_js_1.PublicKey(authorityAddress));
    return [closeAssociatedTokenAccountInstruction];
}
/**
 * Construct Staking Account Authorize Solana instructions
 *
 * @param {StakingAuthorize} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Staking Account Authorize instructions
 */
function stakingAuthorizeInstruction(data) {
    const { params: { stakingAddress, oldAuthorizeAddress, newAuthorizeAddress, newWithdrawAddress }, } = data;
    (0, assert_1.default)(stakingAddress, 'Missing stakingAddress param');
    (0, assert_1.default)(oldAuthorizeAddress, 'Missing oldAuthorizeAddress param');
    (0, assert_1.default)(newAuthorizeAddress, 'Missing newAuthorizeAddress param');
    (0, assert_1.default)(newWithdrawAddress, 'Missing newWithdrawAddress param');
    const tx = new web3_js_1.Transaction();
    const authorizeStaking = web3_js_1.StakeProgram.authorize({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(oldAuthorizeAddress),
        newAuthorizedPubkey: new web3_js_1.PublicKey(newAuthorizeAddress),
        stakeAuthorizationType: web3_js_1.StakeAuthorizationLayout.Staker,
    });
    const authorizeWithdraw = web3_js_1.StakeProgram.authorize({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(oldAuthorizeAddress),
        newAuthorizedPubkey: new web3_js_1.PublicKey(newAuthorizeAddress),
        stakeAuthorizationType: web3_js_1.StakeAuthorizationLayout.Withdrawer,
        custodianPubkey: new web3_js_1.PublicKey(newWithdrawAddress),
    });
    tx.add(authorizeStaking);
    tx.add(authorizeWithdraw);
    return tx.instructions;
}
/**
 * Construct Delegate Solana instructions
 *
 * @param {StakingActivate} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Delegate Solana instructions
 */
function stakingDelegateInstruction(data) {
    const { params: { fromAddress, stakingAddress, validator }, } = data;
    (0, assert_1.default)(fromAddress, 'Missing fromAddress param');
    (0, assert_1.default)(stakingAddress, 'Missing stakingAddress param');
    (0, assert_1.default)(validator, 'Missing validator param');
    const tx = new web3_js_1.Transaction();
    const delegateStaking = web3_js_1.StakeProgram.delegate({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
        votePubkey: new web3_js_1.PublicKey(validator),
    });
    tx.add(delegateStaking);
    return tx.instructions;
}
/**
 * Construct MintTo Solana instructions
 *
 * @param {MintTo} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing MintTo Solana instructions
 */
function mintToInstruction(data) {
    const { params: { mintAddress, destinationAddress, authorityAddress, amount, programId }, } = data;
    (0, assert_1.default)(mintAddress, 'Missing mintAddress param');
    (0, assert_1.default)(destinationAddress, 'Missing destinationAddress param');
    (0, assert_1.default)(authorityAddress, 'Missing authorityAddress param');
    (0, assert_1.default)(amount, 'Missing amount param');
    const mint = new web3_js_1.PublicKey(mintAddress);
    const destination = new web3_js_1.PublicKey(destinationAddress);
    const authority = new web3_js_1.PublicKey(authorityAddress);
    let mintToInstr;
    if (programId && programId === spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
        mintToInstr = (0, spl_token_1.createMintToInstruction)(mint, destination, authority, BigInt(amount), [], spl_token_1.TOKEN_2022_PROGRAM_ID);
    }
    else {
        mintToInstr = (0, spl_token_1.createMintToInstruction)(mint, destination, authority, BigInt(amount));
    }
    return [mintToInstr];
}
/**
 * Construct Burn Solana instructions
 *
 * @param {Burn} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Burn Solana instructions
 */
function burnInstruction(data) {
    const { params: { mintAddress, accountAddress, authorityAddress, amount, programId }, } = data;
    (0, assert_1.default)(mintAddress, 'Missing mintAddress param');
    (0, assert_1.default)(accountAddress, 'Missing accountAddress param');
    (0, assert_1.default)(authorityAddress, 'Missing authorityAddress param');
    (0, assert_1.default)(amount, 'Missing amount param');
    const mint = new web3_js_1.PublicKey(mintAddress);
    const account = new web3_js_1.PublicKey(accountAddress);
    const authority = new web3_js_1.PublicKey(authorityAddress);
    let burnInstr;
    if (programId && programId === spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
        burnInstr = (0, spl_token_1.createBurnInstruction)(account, mint, authority, BigInt(amount), [], spl_token_1.TOKEN_2022_PROGRAM_ID);
    }
    else {
        burnInstr = (0, spl_token_1.createBurnInstruction)(account, mint, authority, BigInt(amount));
    }
    return [burnInstr];
}
/**
 * Process custom instruction - converts to TransactionInstruction
 * Handles conversion from string-based format to TransactionInstruction format
 *
 * @param {CustomInstruction} data - the data containing the custom instruction
 * @returns {TransactionInstruction[]} An array containing the custom instruction
 */
function customInstruction(data) {
    const { params } = data;
    (0, assert_1.default)(params.programId, 'Missing programId in custom instruction');
    (0, assert_1.default)(params.keys && Array.isArray(params.keys), 'Missing or invalid keys in custom instruction');
    (0, assert_1.default)(params.data !== undefined, 'Missing data in custom instruction');
    // Convert string data to Buffer
    let dataBuffer;
    if ((0, utils_1.isValidBase64)(params.data)) {
        dataBuffer = Buffer.from(params.data, 'base64');
    }
    else if ((0, utils_1.isValidHex)(params.data)) {
        dataBuffer = Buffer.from(params.data, 'hex');
    }
    else {
        // Fallback to UTF-8
        dataBuffer = Buffer.from(params.data, 'utf8');
    }
    // Create a new TransactionInstruction with the converted data
    const convertedInstruction = new web3_js_1.TransactionInstruction({
        programId: new web3_js_1.PublicKey(params.programId),
        keys: params.keys.map((key) => ({
            pubkey: new web3_js_1.PublicKey(key.pubkey),
            isSigner: key.isSigner,
            isWritable: key.isWritable,
        })),
        data: dataBuffer,
    });
    return [convertedInstruction];
}
function upsertAccountMeta(keys, meta) {
    const existing = keys.find((account) => account.pubkey.equals(meta.pubkey));
    if (existing) {
        existing.isWritable = existing.isWritable || meta.isWritable;
        existing.isSigner = existing.isSigner || meta.isSigner;
    }
    else {
        keys.push(meta);
    }
}
function buildStaticTransferHookAccounts(transferHook) {
    const metas = [];
    if (transferHook.extraAccountMetas?.length) {
        for (const meta of transferHook.extraAccountMetas) {
            metas.push({
                pubkey: new web3_js_1.PublicKey(meta.pubkey),
                isSigner: meta.isSigner,
                isWritable: meta.isWritable,
            });
        }
    }
    return metas;
}
function addTransferHookAccounts(instruction, transferHook) {
    const extraMetas = buildStaticTransferHookAccounts(transferHook);
    for (const meta of extraMetas) {
        upsertAccountMeta(instruction.keys, meta);
    }
}
