// @bitgo/sdk-coin-sol/dist/cjs/src/lib/tokenTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("./utils");
const constants_1 = require("./constants");
const assert_1 = __importDefault(require("assert"));
const transactionBuilder_1 = require("./transactionBuilder");
const lodash_1 = __importDefault(require("lodash"));
const UNSIGNED_BIGINT_MAX = BigInt('18446744073709551615');
class TokenTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._sendParams = [];
        this._createAtaParams = [];
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.TokenTransfer) {
                const transferInstruction = instruction;
                this.sender(transferInstruction.params.fromAddress);
                this.send({
                    address: transferInstruction.params.toAddress,
                    amount: transferInstruction.params.amount,
                    tokenName: transferInstruction.params.tokenName,
                    tokenAddress: transferInstruction.params.tokenAddress,
                    programId: transferInstruction.params.programId,
                    decimalPlaces: transferInstruction.params.decimalPlaces,
                });
            }
            if (instruction.type === constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount) {
                const ataInitInstruction = instruction;
                this._createAtaParams.push({
                    ownerAddress: ataInitInstruction.params.ownerAddress,
                    tokenName: ataInitInstruction.params.tokenName,
                    ataAddress: ataInitInstruction.params.ataAddress,
                    tokenAddress: ataInitInstruction.params.mintAddress,
                    programId: ataInitInstruction.params.programId,
                });
            }
        }
    }
    /**
     *  Set a transfer
     *
     * @param {SendParams} params - params for the transfer
     * @param {string} params.address - the receiver token address
     * @param {string} params.amount - the amount sent
     * @param {string} params.tokenName - name of token that is intended to send
     * @returns {TransactionBuilder} This transaction builder
     */
    send({ address, amount, tokenName, tokenAddress, programId, decimalPlaces }) {
        (0, utils_1.validateAddress)(address, 'address');
        if (!amount || !(0, utils_1.isValidAmount)(amount)) {
            throw new sdk_core_1.BuildTransactionError('Invalid or missing amount, got: ' + amount);
        }
        if (BigInt(amount) > UNSIGNED_BIGINT_MAX) {
            throw new sdk_core_1.BuildTransactionError(`input amount ${amount} exceeds big int limit ${UNSIGNED_BIGINT_MAX}`);
        }
        this._sendParams.push({ address, amount, tokenName: tokenName, tokenAddress, programId, decimalPlaces });
        return this;
    }
    /**
     *
     * @param {TokenAssociateRecipient} recipient - recipient of the associated token account creation
     * @param {string} recipient.ownerAddress - owner of the associated token account
     * @param {string} recipient.tokenName - name of the token that is intended to associate
     * @returns {TransactionBuilder} This transaction builder
     */
    createAssociatedTokenAccount(recipient) {
        (0, utils_1.validateOwnerAddress)(recipient.ownerAddress);
        const token = (0, utils_1.getSolTokenFromTokenName)(recipient.tokenName);
        let tokenAddress;
        if (recipient.tokenAddress) {
            tokenAddress = recipient.tokenAddress;
        }
        else if (token) {
            tokenAddress = token.tokenAddress;
        }
        else {
            throw new sdk_core_1.BuildTransactionError('Invalid token name, got: ' + recipient.tokenName);
        }
        (0, utils_1.validateMintAddress)(tokenAddress);
        this._createAtaParams.push(recipient);
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        (0, assert_1.default)(this._sender, 'Sender must be set before building the transaction');
        const sendInstructions = await Promise.all(this._sendParams.map(async (sendParams) => {
            const coin = (0, utils_1.getSolTokenFromTokenName)(sendParams.tokenName);
            let tokenAddress;
            let tokenName;
            let programId;
            let decimals;
            if (sendParams.tokenAddress && sendParams.programId && sendParams.decimalPlaces) {
                tokenAddress = sendParams.tokenAddress;
                tokenName = sendParams.tokenName;
                programId = sendParams.programId;
                decimals = sendParams.decimalPlaces;
            }
            else if (coin) {
                tokenAddress = coin.tokenAddress;
                tokenName = coin.name;
                programId = coin.programId;
                decimals = coin.decimalPlaces;
            }
            else {
                throw new Error(`Could not determine token information for ${sendParams.tokenName}`);
            }
            const sourceAddress = await (0, utils_1.getAssociatedTokenAccountAddress)(tokenAddress, this._sender, false, programId);
            return {
                type: constants_1.InstructionBuilderTypes.TokenTransfer,
                params: {
                    fromAddress: this._sender,
                    toAddress: sendParams.address,
                    amount: sendParams.amount,
                    tokenName: tokenName,
                    sourceAddress: sourceAddress,
                    tokenAddress: tokenAddress,
                    programId: programId,
                    decimalPlaces: decimals,
                },
            };
        }));
        const uniqueCreateAtaParams = lodash_1.default.uniqBy(this._createAtaParams, (recipient) => {
            return recipient.ownerAddress + recipient.tokenName;
        });
        const createAtaInstructions = await Promise.all(uniqueCreateAtaParams.map(async (recipient) => {
            const coin = (0, utils_1.getSolTokenFromTokenName)(recipient.tokenName);
            let tokenAddress;
            let tokenName;
            let programId;
            if (coin) {
                tokenName = coin.name;
                tokenAddress = coin.tokenAddress;
                programId = coin.programId;
            }
            else if (recipient.tokenAddress && recipient.programId) {
                tokenName = recipient.tokenName;
                tokenAddress = recipient.tokenAddress;
                programId = recipient.programId;
            }
            else {
                throw new Error(`Could not determine token information for ${recipient.tokenName}`);
            }
            // Use the provided ataAddress if it exists, otherwise calculate it
            let ataAddress = recipient.ataAddress;
            if (!ataAddress) {
                ataAddress = await (0, utils_1.getAssociatedTokenAccountAddress)(tokenAddress, recipient.ownerAddress, false, programId);
            }
            return {
                type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                params: {
                    ownerAddress: recipient.ownerAddress,
                    mintAddress: tokenAddress,
                    ataAddress,
                    payerAddress: this._sender,
                    tokenName: tokenName,
                    programId: programId,
                },
            };
        }));
        const addPriorityFeeInstruction = {
            type: constants_1.InstructionBuilderTypes.SetPriorityFee,
            params: {
                fee: this._priorityFee,
            },
        };
        if (!this._priorityFee || this._priorityFee === Number(0)) {
            this._instructionsData = [...createAtaInstructions, ...sendInstructions];
        }
        else {
            // order is important, createAtaInstructions must be before sendInstructions
            this._instructionsData = [addPriorityFeeInstruction, ...createAtaInstructions, ...sendInstructions];
        }
        return await super.buildImplementation();
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
