// @bitgo/sdk-coin-sol/dist/cjs/src/lib/transferBuilderV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilderV2 = void 0;
const transactionBuilder_1 = require("./transactionBuilder");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("./utils");
const assert_1 = __importDefault(require("assert"));
const constants_1 = require("./constants");
const lodash_1 = __importDefault(require("lodash"));
const UNSIGNED_BIGINT_MAX = BigInt('18446744073709551615');
class TransferBuilderV2 extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._sendParams = [];
        this._createAtaParams = [];
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.Transfer) {
                const transferInstruction = instruction;
                this.sender(transferInstruction.params.fromAddress);
                this.send({
                    address: transferInstruction.params.toAddress,
                    amount: transferInstruction.params.amount,
                });
            }
            else if (instruction.type === constants_1.InstructionBuilderTypes.TokenTransfer) {
                const transferInstruction = instruction;
                this.sender(transferInstruction.params.fromAddress);
                this.send({
                    address: transferInstruction.params.toAddress,
                    amount: transferInstruction.params.amount,
                    tokenName: transferInstruction.params.tokenName,
                });
            }
            else if (instruction.type === constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount) {
                const ataInitInstruction = instruction;
                this._createAtaParams.push({
                    ownerAddress: ataInitInstruction.params.ownerAddress,
                    tokenName: ataInitInstruction.params.tokenName,
                    ataAddress: ataInitInstruction.params.ataAddress,
                });
            }
        }
    }
    /**
     * Set a feePayer
     * @param payerAddress
     */
    feePayer(payerAddress) {
        (0, utils_1.validateAddress)(payerAddress, 'address');
        this._feePayer = payerAddress;
        return this;
    }
    /**
     *  Set a transfer
     *
     * @param {SendParams} sendParams - sendParams
     * @returns {TransactionBuilder} This transaction builder
     */
    send(sendParams) {
        (0, utils_1.validateAddress)(sendParams.address, 'address');
        if (!sendParams.amount || !(0, utils_1.isValidAmount)(sendParams.amount)) {
            throw new sdk_core_1.BuildTransactionError('Invalid or missing amount, got: ' + sendParams.amount);
        }
        if (sendParams.tokenName && BigInt(sendParams.amount) > UNSIGNED_BIGINT_MAX) {
            throw new sdk_core_1.BuildTransactionError(`input amount ${sendParams.amount} exceeds big int limit ${UNSIGNED_BIGINT_MAX}`);
        }
        else if (!sendParams.tokenName && BigInt(sendParams.amount) > BigInt(Number.MAX_SAFE_INTEGER)) {
            throw new sdk_core_1.BuildTransactionError(`input amount ${sendParams.amount} exceeds max safe int ${Number.MAX_SAFE_INTEGER}`);
        }
        this._sendParams.push(sendParams);
        return this;
    }
    /**
     *
     * @param {TokenAssociateRecipient} recipient - recipient of the associated token account creation
     * @param {string} recipient.ownerAddress - owner of the associated token account
     * @param {string} recipient.tokenName - name of the token that is intended to associate
     * @returns {TransactionBuilder} This transaction builder
     */
    createAssociatedTokenAccount(recipient) {
        (0, utils_1.validateOwnerAddress)(recipient.ownerAddress);
        const token = (0, utils_1.getSolTokenFromTokenName)(recipient.tokenName);
        let tokenAddress;
        if (recipient.tokenAddress) {
            tokenAddress = recipient.tokenAddress;
        }
        else if (token) {
            tokenAddress = token.tokenAddress;
        }
        if (!tokenAddress) {
            throw new sdk_core_1.BuildTransactionError('Invalid token name, got: ' + recipient.tokenName);
        }
        (0, utils_1.validateMintAddress)(tokenAddress);
        this._createAtaParams.push(recipient);
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        (0, assert_1.default)(this._sender, 'Sender must be set before building the transaction');
        const sendInstructions = await Promise.all(this._sendParams.map(async (sendParams) => {
            if (sendParams.tokenName) {
                const coin = (0, utils_1.getSolTokenFromTokenName)(sendParams.tokenName);
                let tokenAddress;
                let tokenName;
                let programId;
                let decimals;
                if (sendParams.tokenAddress && sendParams.programId && sendParams.decimalPlaces) {
                    tokenName = sendParams.tokenName;
                    tokenAddress = sendParams.tokenAddress;
                    decimals = sendParams.decimalPlaces;
                    programId = sendParams.programId;
                }
                else if (coin) {
                    tokenName = coin.name;
                    tokenAddress = coin.tokenAddress;
                    decimals = coin.decimalPlaces;
                    programId = coin.programId;
                }
                else {
                    throw new Error(`Could not determine token information for ${sendParams.tokenName}`);
                }
                const sourceAddress = await (0, utils_1.getAssociatedTokenAccountAddress)(tokenAddress, this._sender, false, programId);
                return {
                    type: constants_1.InstructionBuilderTypes.TokenTransfer,
                    params: {
                        fromAddress: this._sender,
                        toAddress: sendParams.address,
                        amount: sendParams.amount,
                        tokenName: tokenName,
                        sourceAddress: sourceAddress,
                        tokenAddress: tokenAddress,
                        programId: programId,
                        decimalPlaces: decimals,
                    },
                };
            }
            else {
                return {
                    type: constants_1.InstructionBuilderTypes.Transfer,
                    params: {
                        fromAddress: this._sender,
                        toAddress: sendParams.address,
                        amount: sendParams.amount,
                    },
                };
            }
        }));
        const uniqueCreateAtaParams = lodash_1.default.uniqBy(this._createAtaParams, (recipient) => {
            return recipient.ownerAddress + recipient.tokenName;
        });
        const createAtaInstructions = await Promise.all(uniqueCreateAtaParams.map(async (recipient) => {
            const coin = (0, utils_1.getSolTokenFromTokenName)(recipient.tokenName);
            let tokenAddress;
            let tokenName;
            let programId;
            if (recipient.tokenAddress && recipient.programId) {
                tokenName = recipient.tokenName;
                tokenAddress = recipient.tokenAddress;
                programId = recipient.programId;
            }
            else if (coin) {
                tokenName = coin.name;
                tokenAddress = coin.tokenAddress;
                programId = coin.programId;
            }
            else {
                throw new Error(`Could not determine token information for ${recipient.tokenName}`);
            }
            const recipientTokenAddress = await (0, utils_1.getAssociatedTokenAccountAddress)(tokenAddress, recipient.ownerAddress, false, programId);
            return {
                type: constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount,
                params: {
                    ownerAddress: recipient.ownerAddress,
                    tokenName: tokenName,
                    mintAddress: tokenAddress,
                    ataAddress: recipientTokenAddress,
                    payerAddress: this._sender,
                    programId: programId,
                },
            };
        }));
        let addPriorityFeeInstruction;
        // If there are createAtaInstructions, then token is involved and we need to add a priority fee instruction
        if (!this._priorityFee || this._priorityFee === Number(0)) {
            this._instructionsData = [...createAtaInstructions, ...sendInstructions];
        }
        else if (createAtaInstructions.length !== 0 ||
            sendInstructions.some((instruction) => instruction.type === constants_1.InstructionBuilderTypes.TokenTransfer)) {
            addPriorityFeeInstruction = {
                type: constants_1.InstructionBuilderTypes.SetPriorityFee,
                params: {
                    fee: this._priorityFee,
                },
            };
            this._instructionsData = [addPriorityFeeInstruction, ...createAtaInstructions, ...sendInstructions];
        }
        return await super.buildImplementation();
    }
}
exports.TransferBuilderV2 = TransferBuilderV2;
