// @bitgo/sdk-coin-sol/dist/cjs/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isIdempotentAtaInstruction = isIdempotentAtaInstruction;
exports.isValidAddress = isValidAddress;
exports.isValidBlockId = isValidBlockId;
exports.isValidPrivateKey = isValidPrivateKey;
exports.isValidPublicKey = isValidPublicKey;
exports.isValidSignature = isValidSignature;
exports.isValidTransactionId = isValidTransactionId;
exports.isValidAmount = isValidAmount;
exports.isValidStakingAmount = isValidStakingAmount;
exports.isValidMemo = isValidMemo;
exports.isValidBase64 = isValidBase64;
exports.isValidHex = isValidHex;
exports.isValidRawTransaction = isValidRawTransaction;
exports.verifySignature = verifySignature;
exports.base58ToUint8Array = base58ToUint8Array;
exports.Uint8ArrayTobase58 = Uint8ArrayTobase58;
exports.countNotNullSignatures = countNotNullSignatures;
exports.requiresAllSignatures = requiresAllSignatures;
exports.matchTransactionTypeByInstructionsOrder = matchTransactionTypeByInstructionsOrder;
exports.getTransactionType = getTransactionType;
exports.getInstructionType = getInstructionType;
exports.validateIntructionTypes = validateIntructionTypes;
exports.validateRawMsgInstruction = validateRawMsgInstruction;
exports.validateRawTransaction = validateRawTransaction;
exports.validateAddress = validateAddress;
exports.getSolTokenFromAddress = getSolTokenFromAddress;
exports.getSolTokenFromAddressOnly = getSolTokenFromAddressOnly;
exports.getSolTokenFromTokenName = getSolTokenFromTokenName;
exports.getAssociatedTokenAccountAddress = getAssociatedTokenAccountAddress;
exports.validateMintAddress = validateMintAddress;
exports.validateOwnerAddress = validateOwnerAddress;
exports.isSolLegacyInstruction = isSolLegacyInstruction;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const assert_1 = __importDefault(require("assert"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const bs58_1 = __importDefault(require("bs58"));
const tweetnacl_1 = __importDefault(require("tweetnacl"));
const constants_1 = require("./constants");
const spl_stake_pool_1 = require("@solana/spl-stake-pool");
const DECODED_BLOCK_HASH_LENGTH = 32; // https://docs.solana.com/developing/programming-model/transactions#blockhash-format
const DECODED_SIGNATURE_LENGTH = 64; // https://docs.solana.com/terminology#signature
const BASE_58_ENCONDING_REGEX = '[1-9A-HJ-NP-Za-km-z]';
const COMPUTE_BUDGET = 'ComputeBudget111111111111111111111111111111';
/**
 * Checks if an ATA instruction is idempotent (discriminator byte = 1).
 *
 * Idempotent ATA instructions use Buffer.from([1]) as instruction data, created by
 * @solana/spl-token's createAssociatedTokenAccountIdempotentInstruction() (v0.4.1+).
 * An idempotent ATA instruction succeed even if account exists,
 * preventing race condition errors in concurrent scenarios.
 *
 * @param {TransactionInstruction} instruction - The instruction to validate
 * @returns {boolean} True if instruction data is a single byte with value 1
 */
function isIdempotentAtaInstruction(instruction) {
    return instruction.data.length === 1 && instruction.data[0] === 1;
}
/** @inheritdoc */
function isValidAddress(address) {
    return isValidPublicKey(address);
}
/** @inheritdoc */
function isValidBlockId(hash) {
    try {
        return (!!hash && new RegExp(BASE_58_ENCONDING_REGEX).test(hash) && bs58_1.default.decode(hash).length === DECODED_BLOCK_HASH_LENGTH);
    }
    catch (e) {
        return false;
    }
}
/** @inheritdoc */
function isValidPrivateKey(prvKey) {
    try {
        const key = typeof prvKey === 'string' ? base58ToUint8Array(prvKey) : prvKey;
        return !!web3_js_1.Keypair.fromSecretKey(key);
    }
    catch (e) {
        return false;
    }
}
/** @inheritdoc */
function isValidPublicKey(pubKey) {
    try {
        if ((0, sdk_core_1.isValidXpub)(pubKey))
            return true;
        new web3_js_1.PublicKey(pubKey);
        return true;
    }
    catch {
        return false;
    }
}
/** @inheritdoc */
function isValidSignature(signature) {
    try {
        return !!signature && bs58_1.default.decode(signature).length === DECODED_SIGNATURE_LENGTH;
    }
    catch (e) {
        return false;
    }
}
/** @inheritdoc */
// TransactionId are the first signature on a Transaction
function isValidTransactionId(txId) {
    return isValidSignature(txId);
}
/**
 * Returns whether or not the string is a valid amount of lamports number
 *
 * @param {string} amount - the string to validate
 * @returns {boolean} - the validation result
 */
function isValidAmount(amount) {
    const bigNumberAmount = new bignumber_js_1.default(amount);
    return bigNumberAmount.isInteger() && bigNumberAmount.isGreaterThanOrEqualTo(0);
}
/**
 * Check if the string is a valid amount of lamports number on staking
 *
 * @param {string} amount - the string to validate
 * @returns {boolean} - the validation result
 */
function isValidStakingAmount(amount) {
    const bigNumberAmount = new bignumber_js_1.default(amount);
    return bigNumberAmount.isInteger() && bigNumberAmount.isGreaterThan(0);
}
/**
 * Check if this is a valid memo or not.
 *
 * @param memo - the memo string
 * @returns {boolean} - the validation result
 */
function isValidMemo(memo) {
    return Buffer.from(memo).length <= constants_1.MAX_MEMO_LENGTH;
}
/**
 * Checks if a string is valid base64 encoded data
 * @param str - The string to validate
 * @returns True if the string is valid base64, false otherwise
 */
function isValidBase64(str) {
    try {
        const decoded = Buffer.from(str, 'base64').toString('base64');
        return decoded === str;
    }
    catch {
        return false;
    }
}
/**
 * Checks if a string is valid hexadecimal data
 * @param str - The string to validate
 * @returns True if the string is valid hex, false otherwise
 */
function isValidHex(str) {
    return /^[0-9A-Fa-f]*$/.test(str) && str.length % 2 === 0;
}
/**
 * Checks if raw transaction can be deserialized
 *
 * @param {string} rawTransaction - transaction in base64 string format
 * @param {boolean} requireAllSignatures - require all signatures to be present
 * @param {boolean} verifySignatures - verify signatures
 * @returns {boolean} - the validation result
 */
function isValidRawTransaction(rawTransaction, requireAllSignatures = false, verifySignatures = false) {
    try {
        const tx = web3_js_1.Transaction.from(Buffer.from(rawTransaction, 'base64'));
        tx.serialize({ requireAllSignatures, verifySignatures });
        return true;
    }
    catch (e) {
        return false;
    }
}
/**
 * Verifies if signature for message is valid.
 *
 * @param {Buffer} serializedTx - tx as a base64 string
 * @param {string} signature - signature as a string
 * @param {string} publicKey - public key as base 58
 * @returns {Boolean} true if signature is valid.
 */
function verifySignature(serializedTx, signature, publicKey) {
    if (!isValidRawTransaction(serializedTx)) {
        throw new sdk_core_1.UtilsError('Invalid serializedTx');
    }
    if (!isValidPublicKey(publicKey)) {
        throw new sdk_core_1.UtilsError('Invalid publicKey');
    }
    if (!isValidSignature(signature)) {
        throw new sdk_core_1.UtilsError('Invalid signature');
    }
    const msg = web3_js_1.Transaction.from(Buffer.from(serializedTx, 'base64')).serializeMessage();
    const sig = base58ToUint8Array(signature);
    const pub = new web3_js_1.PublicKey(publicKey);
    return tweetnacl_1.default.sign.detached.verify(msg, sig, pub.toBuffer());
}
/**
 * Converts a base58 string into a Uint8Array.
 *
 * @param {string} input - a string in base58
 * @returns {Uint8Array} - an Uint8Array
 */
function base58ToUint8Array(input) {
    return new Uint8Array(bs58_1.default.decode(input));
}
/**
 * Converts a Uint8Array to a base58 string.
 *
 * @param {Uint8Array} input - an Uint8Array
 * @returns {string} - a string in base58
 */
function Uint8ArrayTobase58(input) {
    return bs58_1.default.encode(input);
}
/**
 * Count the amount of signatures are not null.
 *
 * @param {SignaturePubkeyPair[]} signatures - an array of SignaturePubkeyPair
 * @returns {number} - the amount of valid signatures
 */
function countNotNullSignatures(signatures) {
    return signatures.filter((sig) => !!sig.signature).length;
}
/**
 * Check if all signatures are completed.
 *
 * @param {SignaturePubkeyPair[]} signatures - signatures
 * @returns {boolean}
 */
function requiresAllSignatures(signatures) {
    return signatures.length > 0 && countNotNullSignatures(signatures) === signatures.length;
}
/**
 * Check the transaction type matching instructions by order. Memo and AdvanceNonceAccount instructions
 * are ignored.
 *
 * @param {TransactionInstruction[]} instructions - the array of supported Solana instructions to be parsed
 * @param {Record<string, number>} instructionIndexes - the instructions indexes of the current transaction
 * @returns true if it matches by order.
 */
function matchTransactionTypeByInstructionsOrder(instructions, instructionIndexes) {
    const instructionsCopy = [...instructions]; // Make a copy since we may modify the array below
    // AdvanceNonceAccount is optional and the first instruction added, it does not matter to match the type
    if (instructionsCopy.length > 0) {
        if (getInstructionType(instructions[0]) === 'AdvanceNonceAccount') {
            instructionsCopy.shift();
        }
    }
    // Memo is optional and the last instruction added, it does not matter to match the type
    // Why have it in instructionKeys if we are going to ignore it?
    const instructionsKeys = Object.keys(instructionIndexes);
    if (instructionsKeys[instructionsKeys.length - 1] === 'Memo') {
        instructionsKeys.pop();
    }
    // Check instructions by order using the index.
    for (const keyName of instructionsKeys) {
        const index = instructionIndexes[keyName];
        if (index >= instructionsCopy.length) {
            return false;
        }
        const result = getInstructionType(instructionsCopy[index]);
        if (result !== keyName) {
            return false;
        }
    }
    return true;
}
/**
 * Returns the transaction Type based on the  transaction instructions.
 * Wallet initialization, Transfer and Staking transactions are supported.
 *
 * @param {SolTransaction} transaction - the solana transaction
 * @returns {TransactionType} - the type of transaction
 */
function getTransactionType(transaction) {
    const { instructions } = transaction;
    if (validateRawMsgInstruction(instructions)) {
        return sdk_core_1.TransactionType.StakingAuthorizeRaw;
    }
    validateIntructionTypes(instructions);
    // check if deactivate instruction does not exist because deactivate can be include a transfer instruction
    const memoInstruction = instructions.find((instruction) => getInstructionType(instruction) === 'Memo');
    const memoData = memoInstruction?.data.toString('utf-8');
    if (memoData?.includes('WalletConnectDefiCustomTx')) {
        return sdk_core_1.TransactionType.CustomTx;
    }
    if (instructions.filter((instruction) => getInstructionType(instruction) === 'Deactivate').length === 0) {
        for (const instruction of instructions) {
            const instructionType = getInstructionType(instruction);
            // Check if memo instruction is there and if it contains 'PrepareForRevoke' because Marinade staking deactivate transaction will have this
            if ((instructionType === constants_1.ValidInstructionTypesEnum.Transfer && !memoData?.includes('PrepareForRevoke')) ||
                instructionType === constants_1.ValidInstructionTypesEnum.TokenTransfer) {
                return sdk_core_1.TransactionType.Send;
            }
        }
    }
    if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.walletInitInstructionIndexes)) {
        return sdk_core_1.TransactionType.WalletInitialization;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.marinadeStakingActivateInstructionsIndexes) ||
        matchTransactionTypeByInstructionsOrder(instructions, constants_1.jitoStakingActivateInstructionsIndexes) ||
        matchTransactionTypeByInstructionsOrder(instructions, constants_1.jitoStakingActivateWithATAInstructionsIndexes) ||
        matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingActivateInstructionsIndexes)) {
        return sdk_core_1.TransactionType.StakingActivate;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingAuthorizeInstructionsIndexes)) {
        return sdk_core_1.TransactionType.StakingAuthorize;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingDelegateInstructionsIndexes)) {
        return sdk_core_1.TransactionType.StakingDelegate;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.marinadeStakingDeactivateInstructionsIndexes) ||
        matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingDeactivateInstructionsIndexes) ||
        matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingPartialDeactivateInstructionsIndexes) ||
        matchTransactionTypeByInstructionsOrder(instructions, constants_1.jitoStakingDeactivateInstructionsIndexes)) {
        return sdk_core_1.TransactionType.StakingDeactivate;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.stakingWithdrawInstructionsIndexes)) {
        return sdk_core_1.TransactionType.StakingWithdraw;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.ataInitInstructionIndexes)) {
        return sdk_core_1.TransactionType.AssociatedTokenAccountInitialization;
    }
    else if (matchTransactionTypeByInstructionsOrder(instructions, constants_1.ataCloseInstructionIndexes)) {
        return sdk_core_1.TransactionType.CloseAssociatedTokenAccount;
    }
    else {
        return sdk_core_1.TransactionType.CustomTx;
    }
}
/**
 * Returns the instruction Type based on the solana instructions.
 * Throws if the solana instruction program is not supported
 *
 * @param {TransactionInstruction} instruction - a solana instruction
 * @returns {ValidInstructionTypes} - a solana instruction type
 */
function getInstructionType(instruction) {
    switch (instruction.programId.toString()) {
        case new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK).toString():
            return 'Memo';
        case web3_js_1.SystemProgram.programId.toString():
            return web3_js_1.SystemInstruction.decodeInstructionType(instruction);
        case spl_token_1.TOKEN_PROGRAM_ID.toString():
        case spl_token_1.TOKEN_2022_PROGRAM_ID.toString():
            const decodedInstruction = (0, spl_token_1.decodeInstruction)(instruction, instruction.programId);
            const instructionTypeMap = new Map();
            instructionTypeMap.set(spl_token_1.TokenInstruction.CloseAccount, 'CloseAssociatedTokenAccount');
            instructionTypeMap.set(spl_token_1.TokenInstruction.Burn, 'Burn');
            instructionTypeMap.set(spl_token_1.TokenInstruction.MintTo, 'MintTo');
            instructionTypeMap.set(spl_token_1.TokenInstruction.Approve, 'Approve');
            instructionTypeMap.set(spl_token_1.TokenInstruction.TransferChecked, 'TokenTransfer');
            const validInstruction = instructionTypeMap.get(decodedInstruction.data.instruction);
            if (!validInstruction) {
                return 'CustomInstruction';
            }
            return validInstruction;
        case spl_stake_pool_1.STAKE_POOL_PROGRAM_ID.toString():
            const discriminator = instruction.data.readUint8(0);
            const layoutKey = Object.entries(spl_stake_pool_1.STAKE_POOL_INSTRUCTION_LAYOUTS).find(([_, v]) => v.index === discriminator)?.[0];
            if (layoutKey === undefined) {
                throw new Error('Instruction type incorrect; unknown discriminator');
            }
            const instructionKey = layoutKey;
            return instructionKey;
        case web3_js_1.StakeProgram.programId.toString():
            return web3_js_1.StakeInstruction.decodeInstructionType(instruction);
        case spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID.toString():
            // TODO: change this when @spl-token supports decoding associated token instructions
            // Support both legacy ATA creation (data.length === 0) and idempotent ATA creation (discriminator = 1)
            // Both instruction types are treated as 'InitializeAssociatedTokenAccount' for compatibility
            if (instruction.data.length === 0 || isIdempotentAtaInstruction(instruction)) {
                return 'InitializeAssociatedTokenAccount';
            }
            else {
                throw new sdk_core_1.NotSupported('Invalid transaction, instruction program id not supported: ' + instruction.programId.toString());
            }
        case COMPUTE_BUDGET:
            if (web3_js_1.ComputeBudgetInstruction.decodeInstructionType(instruction) === 'SetComputeUnitLimit') {
                return 'SetComputeUnitLimit';
            }
            return 'SetPriorityFee';
        default:
            return 'CustomInstruction';
    }
}
/**
 * Validate solana instructions types to see if they are supported by the builder.
 * Throws if the instruction type is invalid.
 *
 * @param {TransactionInstruction} instructions - a solana instruction
 * @returns {void}
 */
function validateIntructionTypes(instructions) {
    for (const instruction of instructions) {
        if (!constants_1.VALID_SYSTEM_INSTRUCTION_TYPES.includes(getInstructionType(instruction))) {
            throw new sdk_core_1.NotSupported('Invalid transaction, instruction type not supported: ' + getInstructionType(instruction));
        }
    }
}
/**
 * Validate solana instructions match raw msg authorize transaction
 *
 * @param {TransactionInstruction} instructions - a solana instruction
 * @returns {boolean} true if the instructions match the raw msg authorize transaction
 */
function validateRawMsgInstruction(instructions) {
    // as web3.js cannot decode authorize instruction from CLI, we need to check it manually first
    if (instructions.length === 2) {
        const programId1 = instructions[0].programId.toString();
        const programId2 = instructions[1].programId.toString();
        if (programId1 === web3_js_1.SystemProgram.programId.toString() && programId2 === web3_js_1.StakeProgram.programId.toString()) {
            const instructionName1 = web3_js_1.SystemInstruction.decodeInstructionType(instructions[0]);
            const data = instructions[1].data.toString('hex');
            if (instructionName1 === constants_1.nonceAdvanceInstruction &&
                (data === constants_1.validInstructionData || data === constants_1.validInstructionData2)) {
                return true;
            }
        }
    }
    if (instructions.length === 3) {
        const programId1 = instructions[0].programId.toString();
        const programId2 = instructions[1].programId.toString();
        const programId3 = instructions[2].programId.toString();
        if (programId1 === web3_js_1.SystemProgram.programId.toString() &&
            programId2 === web3_js_1.StakeProgram.programId.toString() &&
            programId3 === web3_js_1.StakeProgram.programId.toString()) {
            const instructionName1 = web3_js_1.SystemInstruction.decodeInstructionType(instructions[0]);
            const data = instructions[1].data.toString('hex');
            const data2 = instructions[2].data.toString('hex');
            if (instructionName1 === constants_1.nonceAdvanceInstruction &&
                (data === constants_1.validInstructionData || data === constants_1.validInstructionData2) &&
                (data2 === constants_1.validInstructionData || data2 === constants_1.validInstructionData2)) {
                return true;
            }
        }
    }
    return false;
}
/**
 * Check the raw transaction has a valid format in the blockchain context, throw otherwise.
 *
 * @param {string} rawTransaction - Transaction in base64 string  format
 */
function validateRawTransaction(rawTransaction, requireAllSignatures = false, verifySignatures = false) {
    if (!rawTransaction) {
        throw new sdk_core_1.ParseTransactionError('Invalid raw transaction: Undefined');
    }
    if (!isValidRawTransaction(rawTransaction, requireAllSignatures, verifySignatures)) {
        throw new sdk_core_1.ParseTransactionError('Invalid raw transaction');
    }
}
/**
 * Validates address to check if it exists and is a valid Solana public key
 *
 * @param {string} address The address to be validated
 * @param {string} fieldName Name of the field to validate, its needed to return which field is failing on case of error.
 */
function validateAddress(address, fieldName) {
    if (!address || !isValidPublicKey(address)) {
        throw new sdk_core_1.BuildTransactionError(`Invalid or missing ${fieldName}, got: ${address}`);
    }
}
/**
 * Get the statics coin object matching a given Solana token address if it exists
 *
 * @param tokenAddress The token address to match against
 * @param network Solana Mainnet or Testnet
 * @returns statics BaseCoin object for the matching token
 */
function getSolTokenFromAddress(tokenAddress, network) {
    const tokens = statics_1.coins.filter((coin) => {
        if (coin instanceof statics_1.SolCoin) {
            return coin.network.type === network.type && coin.tokenAddress.toLowerCase() === tokenAddress.toLowerCase();
        }
        return false;
    });
    const tokensArray = tokens.map((token) => token);
    if (tokensArray.length >= 1) {
        // there should never be two tokens with the same contract address, so we assert that here
        (0, assert_1.default)(tokensArray.length === 1);
        return tokensArray[0];
    }
    return undefined;
}
/**
 * Get the statics coin object matching a given Solana token address if it exists
 *
 * @param tokenAddress The token address to match against
 * @param network Solana Mainnet or Testnet
 * @returns statics BaseCoin object for the matching token
 */
function getSolTokenFromAddressOnly(tokenAddress) {
    const tokens = statics_1.coins.filter((coin) => {
        if (coin instanceof statics_1.SolCoin) {
            return coin.tokenAddress.toLowerCase() === tokenAddress.toLowerCase();
        }
        return false;
    });
    const tokensArray = tokens.map((token) => token);
    if (tokensArray.length >= 1) {
        // there should never be two tokens with the same contract address, so we assert that here
        (0, assert_1.default)(tokensArray.length === 1);
        return tokensArray[0];
    }
    return undefined;
}
/**
 * Get the solana token object from token name
 * @param tokenName The token name to match against
 * */
function getSolTokenFromTokenName(tokenName) {
    try {
        const token = statics_1.coins.get(tokenName);
        if (!(token.isToken && token instanceof statics_1.SolCoin)) {
            return undefined;
        }
        return token;
    }
    catch (e) {
        if (!(e instanceof statics_1.CoinNotDefinedError)) {
            throw e;
        }
        return undefined;
    }
}
/**
 * Get the solana associated token account address
 * @param tokenAddress token mint address
 * @param ownerAddress The owner of the associated token account
 * @returns The associated token account address
 * */
async function getAssociatedTokenAccountAddress(tokenMintAddress, ownerAddress, allowOwnerOffCurve = false, programId) {
    const mintPublicKey = new web3_js_1.PublicKey(tokenMintAddress);
    const ownerPublicKey = new web3_js_1.PublicKey(ownerAddress);
    // tokenAddress are not on ed25519 curve, so they can't be used as ownerAddress
    if (!allowOwnerOffCurve && !web3_js_1.PublicKey.isOnCurve(ownerPublicKey.toBuffer())) {
        throw new sdk_core_1.UtilsError('Invalid ownerAddress - address off ed25519 curve, got: ' + ownerAddress);
    }
    if (!programId) {
        const coin = getSolTokenFromAddressOnly(tokenMintAddress);
        if (coin && coin instanceof statics_1.SolCoin && 'programId' in coin && coin.programId) {
            programId = coin.programId.toString();
        }
        else {
            programId = spl_token_1.TOKEN_PROGRAM_ID.toString();
        }
    }
    let ataAddress;
    if (programId === spl_token_1.TOKEN_2022_PROGRAM_ID.toString()) {
        ataAddress = await (0, spl_token_1.getAssociatedTokenAddress)(mintPublicKey, ownerPublicKey, allowOwnerOffCurve, spl_token_1.TOKEN_2022_PROGRAM_ID);
    }
    else {
        ataAddress = await (0, spl_token_1.getAssociatedTokenAddress)(mintPublicKey, ownerPublicKey, allowOwnerOffCurve);
    }
    return ataAddress.toString();
}
function validateMintAddress(mintAddress) {
    if (!mintAddress || !isValidAddress(mintAddress)) {
        throw new sdk_core_1.BuildTransactionError('Invalid or missing mintAddress, got: ' + mintAddress);
    }
}
function validateOwnerAddress(ownerAddress) {
    if (!ownerAddress || !isValidAddress(ownerAddress)) {
        throw new sdk_core_1.BuildTransactionError('Invalid or missing ownerAddress, got: ' + ownerAddress);
    }
}
/**
 * Type predicate to check if an instruction is a legacy SolInstruction.
 * Uses 'programId' property as the discriminator since it's unique to legacy instructions
 * Versioned instructions have 'programIdIndex' property.
 *
 * @param instruction - The instruction to check
 * @returns True if the instruction is a SolInstruction (has programId field)
 */
function isSolLegacyInstruction(instruction) {
    return 'programId' in instruction;
}
