// @bitgo/sdk-core/dist/src/account-lib/mpc/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BIP32 = exports.ShamirSecret = exports.Ed25519BIP32 = exports.rangeProof = exports.EDDSA = exports.Eddsa = exports.ECDSA = exports.Ecdsa = void 0;
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
Object.defineProperty(exports, "Ed25519BIP32", { enumerable: true, get: function () { return sdk_lib_mpc_1.Ed25519Bip32HdTree; } });
Object.defineProperty(exports, "BIP32", { enumerable: true, get: function () { return sdk_lib_mpc_1.Secp256k1Bip32HdTree; } });
const shamir_1 = __importDefault(require("./shamir"));
exports.ShamirSecret = shamir_1.default;
__exportStar(require("./curves"), exports);
__exportStar(require("./util"), exports);
var tss_1 = require("./tss");
Object.defineProperty(exports, "Ecdsa", { enumerable: true, get: function () { return tss_1.Ecdsa; } });
Object.defineProperty(exports, "ECDSA", { enumerable: true, get: function () { return tss_1.ECDSA; } });
Object.defineProperty(exports, "Eddsa", { enumerable: true, get: function () { return tss_1.Eddsa; } });
Object.defineProperty(exports, "EDDSA", { enumerable: true, get: function () { return tss_1.EDDSA; } });
Object.defineProperty(exports, "rangeProof", { enumerable: true, get: function () { return tss_1.rangeProof; } });
