// bitgo/dist/test/unit/decryptKeychain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("should");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_api_1 = require("@bitgo/sdk-api");
describe('decryptKeychainPrivateKey', () => {
    const bitgo = new sdk_api_1.BitGoAPI();
    const prv1 = Math.random().toString();
    const password1 = Math.random().toString();
    const prv2 = Math.random().toString();
    const password2 = Math.random().toString();
    it('should decrypt encryptedPrv', () => {
        const keychain = {
            encryptedPrv: bitgo.encrypt({ input: prv1, password: password1 }),
        };
        (0, sdk_core_1.decryptKeychainPrivateKey)(bitgo, keychain, password1).should.equal(prv1);
    });
    it('should decrypt webauthnDevices encryptedPrv', () => {
        const keychain = {
            webauthnDevices: [
                {
                    otpDeviceId: '123',
                    authenticatorInfo: {
                        credID: 'credID',
                        fmt: 'packed',
                        publicKey: 'some value',
                    },
                    prfSalt: '456',
                    encryptedPrv: bitgo.encrypt({ input: prv2, password: password2 }),
                },
            ],
        };
        (0, sdk_core_1.decryptKeychainPrivateKey)(bitgo, keychain, password2).should.equal(prv2);
    });
    it('should try and decrypt all encryptedPrvs', () => {
        const keychain = {
            encryptedPrv: bitgo.encrypt({ input: prv1, password: password1 }),
            webauthnDevices: [
                {
                    otpDeviceId: '123',
                    authenticatorInfo: {
                        credID: 'credID',
                        fmt: 'packed',
                        publicKey: 'some value',
                    },
                    prfSalt: '456',
                    encryptedPrv: bitgo.encrypt({ input: prv2, password: password2 }),
                },
            ],
        };
        (0, sdk_core_1.decryptKeychainPrivateKey)(bitgo, keychain, password2).should.equal(prv2);
    });
    it('should return undefined if no encryptedPrv can be decrypted', () => {
        const keychain = {
            encryptedPrv: bitgo.encrypt({ input: prv1, password: password1 }),
            webauthnDevices: [
                {
                    otpDeviceId: '123',
                    authenticatorInfo: {
                        credID: 'credID',
                        fmt: 'packed',
                        publicKey: 'some value',
                    },
                    prfSalt: '456',
                    encryptedPrv: bitgo.encrypt({ input: prv2, password: password2 }),
                },
            ],
        };
        ((0, sdk_core_1.decryptKeychainPrivateKey)(bitgo, keychain, Math.random().toString()) === undefined).should.equal(true);
    });
    it('should return undefined if no encryptedPrv is present', () => {
        ((0, sdk_core_1.decryptKeychainPrivateKey)(bitgo, {}, 'password') === undefined).should.be.true();
    });
});
