// @bitgo/sdk-coin-coreum/dist/test/unit/coreum.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const sdk_test_1 = require("@bitgo/sdk-test");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const mocha_1 = require("mocha");
const sinon_1 = __importDefault(require("sinon"));
const src_1 = require("../../src");
const constants_1 = require("../../src/lib/constants");
const utils_1 = require("../../src/lib/utils");
const coreum_1 = require("../resources/coreum");
const tcoreum_1 = require("../resources/tcoreum");
const should = require("should");
describe('Coreum', function () {
    let bitgo;
    let coreum;
    let tcoreum;
    let mainnetUtils;
    let testnetUtils;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('coreum', src_1.Coreum.createInstance);
        bitgo.safeRegister('tcoreum', src_1.Tcoreum.createInstance);
        bitgo.initializeTestVars();
        coreum = bitgo.coin('coreum');
        tcoreum = bitgo.coin('tcoreum');
        mainnetUtils = new utils_1.CoreumUtils(statics_1.NetworkType.MAINNET);
        testnetUtils = new utils_1.CoreumUtils(statics_1.NetworkType.TESTNET);
    });
    it('should return the right info', function () {
        coreum.getChain().should.equal('coreum');
        coreum.getFamily().should.equal('coreum');
        coreum.getFullName().should.equal('TX');
        coreum.getBaseFactor().should.equal(1e6);
        tcoreum.getChain().should.equal('tcoreum');
        tcoreum.getFamily().should.equal('coreum');
        tcoreum.getFullName().should.equal('Testnet TX');
        tcoreum.getBaseFactor().should.equal(1e6);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const mainnetAddressDetails = coreum.getAddressDetails(coreum_1.mainnetAddress.noMemoIdAddress);
            mainnetAddressDetails.address.should.equal(coreum_1.mainnetAddress.noMemoIdAddress);
            should.not.exist(mainnetAddressDetails.memoId);
            const testnetAddressDetails = tcoreum.getAddressDetails(tcoreum_1.testnetAddress.noMemoIdAddress);
            testnetAddressDetails.address.should.equal(tcoreum_1.testnetAddress.noMemoIdAddress);
            should.not.exist(testnetAddressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const mainnetAddressDetails = coreum.getAddressDetails(coreum_1.mainnetAddress.validMemoIdAddress);
            mainnetAddressDetails.address.should.equal(coreum_1.mainnetAddress.validMemoIdAddress.split('?')[0]);
            mainnetAddressDetails.memoId.should.equal('2');
            const testnetAddressDetails = coreum.getAddressDetails(tcoreum_1.testnetAddress.validMemoIdAddress);
            testnetAddressDetails.address.should.equal(tcoreum_1.testnetAddress.validMemoIdAddress.split('?')[0]);
            testnetAddressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                coreum.getAddressDetails(coreum_1.mainnetAddress.multipleMemoIdAddress);
            }).should.throw();
            (() => {
                tcoreum.getAddressDetails(tcoreum_1.testnetAddress.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const mainnetReceiveAddress = {
                address: 'core1ssh2d2ft6hzrgn9z6k7mmsamy2hfpxl9y8re5x?memoId=7',
                coinSpecific: {
                    rootAddress: 'core1ssh2d2ft6hzrgn9z6k7mmsamy2hfpxl9y8re5x',
                    memoID: '7',
                },
            };
            const testnetReceiveAddress = {
                address: 'testcore1ecqgwd4whevrzjxrhrja54c5jg043j79xtz5a5?memoId=7',
                coinSpecific: {
                    rootAddress: 'testcore1ecqgwd4whevrzjxrhrja54c5jg043j79xtz5a5',
                    memoID: '7',
                },
            };
            const isValidMainnetReceiveAddress = await coreum.isWalletAddress(mainnetReceiveAddress);
            const isValidTestnetReceiveAddress = await tcoreum.isWalletAddress(testnetReceiveAddress);
            isValidMainnetReceiveAddress.should.equal(true);
            isValidTestnetReceiveAddress.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(mainnetUtils.isValidAddress(coreum_1.mainnetAddress.address1), true);
            should.equal(mainnetUtils.isValidAddress(coreum_1.mainnetAddress.address2), true);
            should.equal(mainnetUtils.isValidAddress(coreum_1.mainnetAddress.address3), false);
            should.equal(mainnetUtils.isValidAddress(coreum_1.mainnetAddress.address4), false);
            should.equal(mainnetUtils.isValidAddress('dfjk35y'), false);
            should.equal(mainnetUtils.isValidAddress(undefined), false);
            should.equal(mainnetUtils.isValidAddress(''), false);
            should.equal(mainnetUtils.isValidAddress(coreum_1.mainnetAddress.validMemoIdAddress), true);
            should.equal(mainnetUtils.isValidAddress(coreum_1.mainnetAddress.multipleMemoIdAddress), false);
            should.equal(testnetUtils.isValidAddress(tcoreum_1.testnetAddress.address1), true);
            should.equal(testnetUtils.isValidAddress(tcoreum_1.testnetAddress.address2), true);
            should.equal(testnetUtils.isValidAddress(tcoreum_1.testnetAddress.address3), false);
            should.equal(testnetUtils.isValidAddress(tcoreum_1.testnetAddress.address4), false);
            should.equal(testnetUtils.isValidAddress('dfjk35y'), false);
            should.equal(testnetUtils.isValidAddress(undefined), false);
            should.equal(testnetUtils.isValidAddress(''), false);
            should.equal(testnetUtils.isValidAddress(tcoreum_1.testnetAddress.validMemoIdAddress), true);
            should.equal(testnetUtils.isValidAddress(tcoreum_1.testnetAddress.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(mainnetUtils.isValidValidatorAddress(coreum_1.mainnetAddress.validatorAddress1), true);
            should.equal(mainnetUtils.isValidValidatorAddress(coreum_1.mainnetAddress.validatorAddress2), true);
            should.equal(mainnetUtils.isValidValidatorAddress(coreum_1.mainnetAddress.validatorAddress3), false);
            should.equal(mainnetUtils.isValidValidatorAddress(coreum_1.mainnetAddress.validatorAddress4), false);
            should.equal(mainnetUtils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(mainnetUtils.isValidValidatorAddress(undefined), false);
            should.equal(mainnetUtils.isValidValidatorAddress(''), false);
            should.equal(testnetUtils.isValidValidatorAddress(tcoreum_1.testnetAddress.validatorAddress1), true);
            should.equal(testnetUtils.isValidValidatorAddress(tcoreum_1.testnetAddress.validatorAddress2), true);
            should.equal(testnetUtils.isValidValidatorAddress(tcoreum_1.testnetAddress.validatorAddress3), false);
            should.equal(testnetUtils.isValidValidatorAddress(tcoreum_1.testnetAddress.validatorAddress4), false);
            should.equal(testnetUtils.isValidValidatorAddress('dfjk35y'), false);
            should.equal(testnetUtils.isValidValidatorAddress(undefined), false);
            should.equal(testnetUtils.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: tcoreum_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: tcoreum_1.TEST_SEND_TX.recipient,
                        amount: tcoreum_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await tcoreum.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify delegate transaction', async function () {
            const txPrebuild = {
                txHex: tcoreum_1.TEST_DELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: tcoreum_1.TEST_DELEGATE_TX.validator,
                        amount: tcoreum_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await tcoreum.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify undelegate transaction', async function () {
            const txPrebuild = {
                txHex: tcoreum_1.TEST_UNDELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: tcoreum_1.TEST_UNDELEGATE_TX.validator,
                        amount: tcoreum_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await tcoreum.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify withdraw rewards transaction', async function () {
            const txPrebuild = {
                txHex: tcoreum_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: tcoreum_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await tcoreum.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await tcoreum
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await tcoreum.explainTransaction({
                txHex: tcoreum_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: tcoreum_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: tcoreum_1.TEST_SEND_TX.recipient,
                        amount: tcoreum_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: tcoreum_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: tcoreum_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a delegate transaction', async function () {
            const explainedTransaction = await tcoreum.explainTransaction({
                txHex: tcoreum_1.TEST_DELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: tcoreum_1.TEST_DELEGATE_TX.hash,
                outputs: [
                    {
                        address: tcoreum_1.TEST_DELEGATE_TX.validator,
                        amount: tcoreum_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: tcoreum_1.TEST_DELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: tcoreum_1.TEST_DELEGATE_TX.gasBudget.amount[0].amount },
                type: 13,
            });
        });
        it('should explain a undelegate transaction', async function () {
            const explainedTransaction = await tcoreum.explainTransaction({
                txHex: tcoreum_1.TEST_UNDELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: tcoreum_1.TEST_UNDELEGATE_TX.hash,
                outputs: [
                    {
                        address: tcoreum_1.TEST_UNDELEGATE_TX.validator,
                        amount: tcoreum_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: tcoreum_1.TEST_UNDELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: tcoreum_1.TEST_UNDELEGATE_TX.gasBudget.amount[0].amount },
                type: 17,
            });
        });
        it('should explain a withdraw transaction', async function () {
            const explainedTransaction = await tcoreum.explainTransaction({
                txHex: tcoreum_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: tcoreum_1.TEST_WITHDRAW_REWARDS_TX.hash,
                outputs: [
                    {
                        address: tcoreum_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
                outputAmount: undefined,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: tcoreum_1.TEST_WITHDRAW_REWARDS_TX.gasBudget.amount[0].amount },
                type: 15,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await tcoreum.explainTransaction({
                txHex: tcoreum_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: tcoreum_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: tcoreum_1.TEST_TX_WITH_MEMO.to,
                        amount: tcoreum_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: tcoreum_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: tcoreum_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: tcoreum_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await tcoreum.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await tcoreum.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        it('should parse a transfer transaction', async function () {
            const transferInputsResponse = {
                address: tcoreum_1.TEST_SEND_TX.recipient,
                amount: new bignumber_js_1.default(tcoreum_1.TEST_SEND_TX.sendAmount).plus(tcoreum_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
            };
            const transferOutputsResponse = {
                address: tcoreum_1.TEST_SEND_TX.recipient,
                amount: tcoreum_1.TEST_SEND_TX.sendAmount,
            };
            const parsedTransaction = await tcoreum.parseTransaction({ txHex: tcoreum_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Coreum.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await tcoreum
                .parseTransaction({ txHex: tcoreum_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
    describe('Recover transaction: success path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = tcoreum_1.wrwUser.destinationAddress;
        const coin = statics_1.coins.get('tcoreum');
        const testBalance = [
            {
                denom: 'utestcore',
                amount: '150000000000',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.Tcoreum.prototype, 'getAccountBalance');
            accountBalance.withArgs(tcoreum_1.wrwUser.senderAddress).resolves(testBalance);
            const accountDetails = sandBox.stub(src_1.Tcoreum.prototype, 'getAccountDetails');
            accountDetails.withArgs(tcoreum_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Tcoreum.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should recover funds for non-bitgo recoveries', async function () {
            const res = await tcoreum.recover({
                userKey: tcoreum_1.wrwUser.userPrivateKey,
                backupKey: tcoreum_1.wrwUser.backupPrivateKey,
                bitgoKey: tcoreum_1.wrwUser.bitgoPublicKey,
                walletPassphrase: tcoreum_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(tcoreum.getAccountBalance);
            sandBox.assert.calledOnce(tcoreum.getAccountDetails);
            sandBox.assert.calledOnce(tcoreum.getChainId);
            const tcoreumTxn = new abstract_cosmos_1.CosmosTransaction(coin, testnetUtils);
            tcoreumTxn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const tcoreumTxnJson = tcoreumTxn.toJson();
            const sendMessage = tcoreumTxnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const actualBalance = balance.minus(new bignumber_js_1.default(constants_1.GAS_AMOUNT));
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
            should.equal(sendMessage.amount[0].denom, testBalance[0].denom);
        });
    });
    describe('Recover transaction: failure path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = tcoreum_1.wrwUser.destinationAddress;
        const testZeroBalance = [
            {
                denom: 'utestcore',
                amount: '0',
            },
        ];
        const testAccountNumber = '123';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        (0, mocha_1.beforeEach)(() => {
            const accountBalance = sandBox.stub(src_1.Tcoreum.prototype, 'getAccountBalance');
            accountBalance.withArgs(tcoreum_1.wrwUser.senderAddress).resolves(testZeroBalance);
            const accountDetails = sandBox.stub(src_1.Tcoreum.prototype, 'getAccountDetails');
            accountDetails.withArgs(tcoreum_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Tcoreum.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(() => {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should throw error if backupkey is not present', async function () {
            await tcoreum
                .recover({
                userKey: tcoreum_1.wrwUser.userPrivateKey,
                bitgoKey: tcoreum_1.wrwUser.bitgoPublicKey,
                walletPassphrase: tcoreum_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing backupKey');
        });
        it('should throw error if userkey is not present', async function () {
            await tcoreum
                .recover({
                backupKey: tcoreum_1.wrwUser.backupPrivateKey,
                bitgoKey: tcoreum_1.wrwUser.bitgoPublicKey,
                walletPassphrase: tcoreum_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing userKey');
        });
        it('should throw error if wallet passphrase is not present', async function () {
            await tcoreum
                .recover({
                userKey: tcoreum_1.wrwUser.userPrivateKey,
                backupKey: tcoreum_1.wrwUser.backupPrivateKey,
                bitgoKey: tcoreum_1.wrwUser.bitgoPublicKey,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing wallet passphrase');
        });
        it('should throw error if there is no balance', async function () {
            await tcoreum
                .recover({
                userKey: tcoreum_1.wrwUser.userPrivateKey,
                backupKey: tcoreum_1.wrwUser.backupPrivateKey,
                bitgoKey: tcoreum_1.wrwUser.bitgoPublicKey,
                walletPassphrase: tcoreum_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('Did not have enough funds to recover');
        });
    });
});
