// @bitgo/sdk-coin-icp/dist/src/lib/icpAgent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IcpAgent = void 0;
const principal_1 = require("@dfinity/principal");
const ic0_1 = require("ic0");
const utils_1 = __importDefault(require("./utils"));
const iface_1 = require("./iface");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const statics_1 = require("@bitgo/statics");
class IcpAgent {
    constructor(host, staticsCoin) {
        this.host = host;
        this.staticsCoin = staticsCoin;
    }
    /**
     * Creates a new HTTP agent for communicating with the Internet Computer.
     * @returns An instance of HttpAgent.
     */
    createAgent() {
        return new ic0_1.HttpAgent({
            host: this.host,
            fetch,
            verifyQuerySignatures: false,
        });
    }
    /**
     * Retrieves an instance of the ledger canister agent.
     *
     * This method creates a new agent using `createAgent()`, initializes a replica interface
     * with the agent (configured for local use), and returns an `AgentCanister` instance
     * for the ledger canister identified by `LEDGER_CANISTER_ID`.
     *
     * @returns {AgentCanister} An agent interface for interacting with the ledger canister.
     */
    getLedger() {
        const agent = this.createAgent();
        const ic = (0, ic0_1.replica)(agent, { local: true });
        const ledgerCanisterId = this.staticsCoin.network.type === statics_1.NetworkType.TESTNET ? iface_1.TESTNET_LEDGER_CANISTER_ID : iface_1.LEDGER_CANISTER_ID;
        return ic(principal_1.Principal.fromUint8Array(ledgerCanisterId).toText());
    }
    /**
     * Fetches the account balance for a given principal ID.
     * @param principalId - The principal ID of the account.
     * @returns Promise resolving to the account balance as a string.
     * @throws Error if the balance could not be fetched.
     */
    async getBalance(principalId) {
        try {
            if (!principalId) {
                throw new Error('Principal ID is required');
            }
            const ledger = this.getLedger();
            const account = {
                owner: principal_1.Principal.fromText(principalId),
                subaccount: [utils_1.default.hexToBytes(iface_1.DEFAULT_SUBACCOUNT)],
            };
            const balance = await ledger.call(iface_1.ACCOUNT_BALANCE_CALL, account);
            return (0, bignumber_js_1.default)(balance);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Error fetching balance for principal ${principalId}: ${errorMessage}`);
        }
    }
    /**
     * Fetches the transaction fee from the ledger.
     * @returns Promise resolving to the transaction fee as a string.
     * @throws Error if the fee could not be fetched.
     */
    async getFee() {
        try {
            const ledger = this.getLedger();
            const metadata = await ledger.call(iface_1.METADATA_CALL);
            const feeEntry = metadata.find((entry) => entry[0] === iface_1.ICRC1_FEE_KEY && entry[1]?.Nat !== undefined);
            if (!feeEntry) {
                throw new Error(`${iface_1.ICRC1_FEE_KEY} metadata not found or invalid format`);
            }
            return (0, bignumber_js_1.default)(feeEntry[1].Nat);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Error fetching transaction fee: ${errorMessage}`);
        }
    }
}
exports.IcpAgent = IcpAgent;
