// @bitgo/sdk-coin-icp/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestNetwork = exports.Network = exports.MethodName = exports.OperationType = exports.CurveType = exports.SignatureType = exports.RequestType = exports.DEFAULT_SUBACCOUNT = exports.METADATA_CALL = exports.ICRC1_FEE_KEY = exports.DEFAULT_MEMO = exports.PUBLIC_NODE_REQUEST_ENDPOINT = exports.ACCOUNT_BALANCE_CALL = exports.ROOT_PATH = exports.TESTNET_LEDGER_CANISTER_ID = exports.LEDGER_CANISTER_ID = exports.PERMITTED_DRIFT = exports.MAX_INGRESS_TTL = void 0;
exports.MAX_INGRESS_TTL = 5 * 60 * 1000000000; // 5 minutes in nanoseconds
exports.PERMITTED_DRIFT = 60 * 1000000000; // 60 seconds in nanoseconds
exports.LEDGER_CANISTER_ID = new Uint8Array([0, 0, 0, 0, 0, 0, 0, 2, 1, 1]); // Uint8Array value for "00000000000000020101" and the string value is "ryjl3-tyaaa-aaaaa-aaaba-cai"
exports.TESTNET_LEDGER_CANISTER_ID = new Uint8Array([0, 0, 0, 0, 1, 0, 130, 251, 1, 1]); // Uint8Array value for "00000000010082fb0101" and the string value is "xafvr-biaaa-aaaai-aql5q-cai"
exports.ROOT_PATH = 'm/0';
exports.ACCOUNT_BALANCE_CALL = 'icrc1_balance_of';
exports.PUBLIC_NODE_REQUEST_ENDPOINT = '/api/v3/canister/';
exports.DEFAULT_MEMO = 0; // default memo value is 0
exports.ICRC1_FEE_KEY = 'icrc1:fee';
exports.METADATA_CALL = 'icrc1_metadata';
exports.DEFAULT_SUBACCOUNT = '0000000000000000000000000000000000000000000000000000000000000000'; // default subaccount value is 0x
var RequestType;
(function (RequestType) {
    RequestType["CALL"] = "call";
})(RequestType || (exports.RequestType = RequestType = {}));
var SignatureType;
(function (SignatureType) {
    SignatureType["ECDSA"] = "ecdsa";
})(SignatureType || (exports.SignatureType = SignatureType = {}));
var CurveType;
(function (CurveType) {
    CurveType["SECP256K1"] = "secp256k1";
})(CurveType || (exports.CurveType = CurveType = {}));
var OperationType;
(function (OperationType) {
    OperationType["TRANSACTION"] = "TRANSACTION";
    OperationType["FEE"] = "FEE";
})(OperationType || (exports.OperationType = OperationType = {}));
var MethodName;
(function (MethodName) {
    MethodName["SEND_PB"] = "send_pb";
})(MethodName || (exports.MethodName = MethodName = {}));
var Network;
(function (Network) {
    Network["ID"] = "00000000000000020101";
})(Network || (exports.Network = Network = {}));
var TestNetwork;
(function (TestNetwork) {
    TestNetwork["ID"] = "00000000010082fb0101";
})(TestNetwork || (exports.TestNetwork = TestNetwork = {}));
