// @bitgo/sdk-coin-icp/dist/src/lib/keyPair.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("./utils"));
const secp256k1_1 = require("@bitgo/secp256k1");
const crypto_1 = require("crypto");
const DEFAULT_SEED_SIZE_BYTES = 16;
/**
 * ICP keys and address management.
 */
class KeyPair extends sdk_core_1.Secp256k1ExtendedKeyPair {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key (extended or raw), or a public key
     *     (extended, compressed, or uncompressed)
     */
    constructor(source) {
        super(source);
        if (!source) {
            const seed = (0, crypto_1.randomBytes)(DEFAULT_SEED_SIZE_BYTES);
            this.hdNode = secp256k1_1.bip32.fromSeed(seed);
        }
        else if ((0, sdk_core_1.isSeed)(source)) {
            this.hdNode = secp256k1_1.bip32.fromSeed(source.seed);
        }
        else if ((0, sdk_core_1.isPrivateKey)(source)) {
            super.recordKeysFromPrivateKey(source.prv);
        }
        else if ((0, sdk_core_1.isPublicKey)(source)) {
            super.recordKeysFromPublicKey(source.pub);
        }
        else {
            throw new Error('Invalid key pair options');
        }
        if (this.hdNode) {
            this.keyPair = sdk_core_1.Secp256k1ExtendedKeyPair.toKeyPair(this.hdNode);
        }
    }
    /** @inheritdoc */
    getKeys() {
        return {
            pub: this.getPublicKey({ compressed: true }).toString('hex'),
            prv: this.getPrivateKey()?.toString('hex'),
        };
    }
    /** @inheritdoc */
    getAddress() {
        const principal = utils_1.default.derivePrincipalFromPublicKey(this.getKeys().pub);
        const subAccount = new Uint8Array(32);
        const accountId = utils_1.default.fromPrincipal(principal, subAccount);
        return accountId.toString();
    }
}
exports.KeyPair = KeyPair;
