// @bitgo/sdk-coin-icp/dist/src/lib/signedTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignedTransactionBuilder = void 0;
const iface_1 = require("./iface");
const utils_1 = __importDefault(require("./utils"));
const assert_1 = __importDefault(require("assert"));
class SignedTransactionBuilder {
    constructor(unsigned_transaction, signatures) {
        this._unsigned_transaction = unsigned_transaction;
        this._signaturePayload = signatures;
    }
    getSignTransaction() {
        const combineRequest = {
            signatures: this._signaturePayload,
            unsigned_transaction: this._unsigned_transaction,
        };
        const signatureMap = new Map();
        for (const sig of combineRequest.signatures) {
            signatureMap.set(sig.signing_payload.hex_bytes, sig);
        }
        const unsignedTransaction = utils_1.default.cborDecode(utils_1.default.blobFromHex(combineRequest.unsigned_transaction));
        // at present we expect only one request type and one signature
        (0, assert_1.default)(combineRequest.signatures.length === 1);
        (0, assert_1.default)(unsignedTransaction.updates.length === 1);
        const envelopes = this.getEnvelopes(unsignedTransaction, signatureMap);
        const requestEnvelopeArray = envelopes[0];
        const requestEnvelopes = requestEnvelopeArray[0];
        (0, assert_1.default)(requestEnvelopes.length === 1);
        const requestEnvelope = requestEnvelopes[0];
        const signedTransaction = utils_1.default.cborEncode(requestEnvelope.update);
        return signedTransaction;
    }
    getEnvelopes(unsignedTransaction, signatureMap) {
        const envelopes = [];
        for (const [, update] of unsignedTransaction.updates) {
            const requestEnvelopes = [];
            if (unsignedTransaction.ingress_expiries.length != 1) {
                throw new Error('ingress expiry can have only one entry');
            }
            const ingressExpiry = unsignedTransaction.ingress_expiries[0];
            update.ingress_expiry = ingressExpiry;
            const transactionSignature = utils_1.default.getTransactionSignature(signatureMap, update);
            if (!transactionSignature) {
                throw new Error('Transaction signature is invalid');
            }
            const pk_der = utils_1.default.getPublicKeyInDERFormat(transactionSignature.public_key.hex_bytes);
            const updateEnvelope = {
                content: { request_type: iface_1.RequestType.CALL, ...update },
                sender_pubkey: pk_der,
                sender_sig: utils_1.default.blobFromHex(transactionSignature.hex_bytes),
            };
            requestEnvelopes.push({
                update: updateEnvelope,
            });
            envelopes.push([requestEnvelopes]);
        }
        return envelopes;
    }
}
exports.SignedTransactionBuilder = SignedTransactionBuilder;
