// @bitgo/sdk-coin-icp/dist/src/lib/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const iface_1 = require("./iface");
const utils_1 = __importDefault(require("./utils"));
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get icpTransactionData() {
        return this._icpTransactionData;
    }
    set icpTransactionData(icpTransactionData) {
        this._icpTransactionData = icpTransactionData;
    }
    get icpTransaction() {
        return this._icpTransaction;
    }
    set icpTransaction(icpTransaction) {
        this._icpTransaction = icpTransaction;
    }
    get unsignedTransaction() {
        return this._payloadsData.unsigned_transaction;
    }
    get signaturePayload() {
        return this._signaturePayload;
    }
    set signedTransaction(signature) {
        this._signedTransaction = signature;
    }
    get signedTransaction() {
        return this._signedTransaction;
    }
    set payloadsData(payloadsData) {
        this._payloadsData = payloadsData;
    }
    get payloadsData() {
        return this._payloadsData;
    }
    set createdTimestamp(createdTimestamp) {
        this._createdTimestamp = createdTimestamp;
    }
    get createdTimestamp() {
        return this._createdTimestamp;
    }
    async fromRawTransaction(rawTransaction) {
        try {
            const serializedTxFormatBuffer = Buffer.from(rawTransaction, 'hex');
            const serializedTxFormatJsonString = serializedTxFormatBuffer.toString('utf-8');
            const jsonRawTransaction = JSON.parse(serializedTxFormatJsonString);
            const payloadsData = jsonRawTransaction.serializedTxHex;
            this._payloadsData = payloadsData;
            const parsedTx = await this.parseUnsignedTransaction(payloadsData.unsigned_transaction);
            const senderPublicKeyHex = jsonRawTransaction.publicKey;
            const transactionType = parsedTx.operations[0].type;
            switch (transactionType) {
                case iface_1.OperationType.TRANSACTION:
                    this._icpTransactionData = {
                        senderAddress: parsedTx.operations[0].account.address,
                        receiverAddress: parsedTx.operations[1].account.address,
                        amount: parsedTx.operations[1].amount.value,
                        fee: parsedTx.operations[2].amount.value,
                        senderPublicKeyHex: senderPublicKeyHex,
                        transactionType: transactionType,
                        expiryTime: Number(parsedTx.metadata.ingress_end ?? parsedTx.metadata.created_at_time + iface_1.MAX_INGRESS_TTL),
                        memo: parsedTx.metadata.memo,
                    };
                    utils_1.default.validateRawTransaction(this._icpTransactionData);
                    this._id = this.generateTransactionId();
                    break;
                default:
                    throw new Error('Invalid transaction type');
            }
        }
        catch (error) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid transaction: ${error.message}`);
        }
    }
    addSignature(signaturePayloads) {
        if (!signaturePayloads) {
            throw new Error('signatures not provided');
        }
        if (signaturePayloads.length !== this._payloadsData.payloads.length) {
            throw new Error('signatures length is not matching');
        }
        this._signaturePayload = signaturePayloads;
        if (this._id === undefined || this._id === null) {
            this._id = this.generateTransactionId();
        }
    }
    /** @inheritdoc */
    toJson() {
        if (!this._icpTransactionData) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        switch (this._icpTransactionData.transactionType) {
            case iface_1.OperationType.TRANSACTION:
                const txData = {
                    id: this._id,
                    sender: this._icpTransactionData.senderAddress,
                    senderPublicKey: this._icpTransactionData.senderPublicKeyHex,
                    recipient: this._icpTransactionData.receiverAddress,
                    memo: this._icpTransactionData.memo,
                    feeAmount: this._icpTransactionData.fee,
                    expirationTime: this._icpTransactionData.expiryTime,
                    type: sdk_core_1.TransactionType.Send,
                };
                if (this._icpTransactionData.memo !== undefined) {
                    txData.memo = this._icpTransactionData.memo;
                }
                return txData;
            default:
                throw new Error(`Unsupported transaction type: ${this._icpTransactionData.transactionType}`);
        }
    }
    /** @inheritDoc */
    explainTransaction() {
        const result = this.toJson();
        const displayOrder = ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee'];
        const outputs = [];
        const explanationResult = {
            displayOrder,
            id: this.id,
            outputs,
            outputAmount: '0',
            fee: { fee: '0' },
            type: result.type,
            changeOutputs: [], // account based does not use change outputs
            changeAmount: '0', // account based does not make change
        };
        switch (explanationResult.type) {
            case sdk_core_1.TransactionType.Send:
                return this.explainTransferTransaction(explanationResult);
            default:
                throw new sdk_core_1.InvalidTransactionError('Transaction type not supported');
        }
    }
    /**
     * Explains a transfer transaction by providing details about the recipients and the total output amount.
     *
     * @param {IcpTransactionExplanation} explanationResult - The initial explanation result to be extended.
     * @returns {IcpTransactionExplanation} The extended explanation result including the output amount and recipients.
     */
    explainTransferTransaction(explanationResult) {
        explanationResult.fee = { fee: this.icpTransactionData.fee };
        const recipients = utils_1.default.getRecipients(this.icpTransactionData);
        const outputs = [recipients];
        const outputAmountBN = recipients.amount;
        const outputAmount = outputAmountBN.toString();
        return {
            ...explanationResult,
            outputAmount,
            outputs,
        };
    }
    /** @inheritdoc */
    toBroadcastFormat() {
        if (!this._signedTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Empty transaction');
        }
        return this.serialize();
    }
    serialize() {
        return this._signedTransaction;
    }
    async parseUnsignedTransaction(rawTransaction) {
        const unsignedTransaction = utils_1.default.cborDecode(utils_1.default.blobFromHex(rawTransaction));
        const update = unsignedTransaction.updates[0];
        const httpCanisterUpdate = update[1];
        httpCanisterUpdate.ingress_expiry = BigInt(unsignedTransaction.ingress_expiries[0]);
        return await this.getParsedTransactionFromUpdate(httpCanisterUpdate, false);
    }
    async getParsedTransactionFromUpdate(httpCanisterUpdate, isSigned) {
        const senderPrincipal = utils_1.default.convertSenderBlobToPrincipal(httpCanisterUpdate.sender);
        const ACCOUNT_ID_PREFIX = utils_1.default.getAccountIdPrefix();
        const subAccount = new Uint8Array(32);
        const senderAccount = utils_1.default.getAccountIdFromPrincipalBytes(ACCOUNT_ID_PREFIX, Buffer.from(senderPrincipal.buffer), subAccount);
        const args = await utils_1.default.fromArgs(httpCanisterUpdate.arg);
        const senderOperation = {
            type: iface_1.OperationType.TRANSACTION,
            account: { address: senderAccount },
            amount: {
                value: `-${args.payment.receiverGets.e8s.toString()}`,
                currency: {
                    symbol: this._coinConfig.family,
                    decimals: this._coinConfig.decimalPlaces,
                },
            },
        };
        const receiverOperation = {
            type: iface_1.OperationType.TRANSACTION,
            account: { address: args.to.hash.toString('hex') },
            amount: {
                value: args.payment.receiverGets.e8s.toString(),
                currency: {
                    symbol: this._coinConfig.family,
                    decimals: this._coinConfig.decimalPlaces,
                },
            },
        };
        const feeOperation = {
            type: iface_1.OperationType.FEE,
            account: { address: senderAccount },
            amount: {
                value: `-${args.maxFee.e8s.toString()}`,
                currency: {
                    symbol: this._coinConfig.family,
                    decimals: this._coinConfig.decimalPlaces,
                },
            },
        };
        const accountIdentifierSigners = [];
        if (isSigned) {
            accountIdentifierSigners.push({ address: senderAccount });
        }
        const parsedTxn = {
            operations: [senderOperation, receiverOperation, feeOperation],
            metadata: {
                created_at_time: args.createdAtTime.timestampNanos,
                memo: Number(args.memo.memo),
                ingress_end: Number(httpCanisterUpdate.ingress_expiry) + iface_1.PERMITTED_DRIFT,
            },
            account_identifier_signers: accountIdentifierSigners,
        };
        this.createdTimestamp = args.createdAtTime.timestampNanos;
        return parsedTxn;
    }
    async parseSignedTransaction(rawTransaction) {
        const signedTransaction = utils_1.default.cborDecode(utils_1.default.blobFromHex(rawTransaction));
        const httpCanisterUpdate = signedTransaction.content;
        httpCanisterUpdate.ingress_expiry = BigInt(signedTransaction.content.ingress_expiry);
        return await this.getParsedTransactionFromUpdate(httpCanisterUpdate, true);
    }
    /** @inheritdoc */
    canSign(key) {
        return true;
    }
    /**
     * Generates a unique transaction ID for the current transaction.
     * The transaction ID is derived using the unsigned transaction data,
     * the sender's address, and the receiver's address.
     *
     * @returns {string} The generated transaction ID.
     */
    generateTransactionId() {
        const id = utils_1.default.getTransactionId(this.unsignedTransaction, this.icpTransactionData.senderAddress, this.icpTransactionData.receiverAddress);
        return id;
    }
}
exports.Transaction = Transaction;
