// @bitgo/sdk-coin-icp/dist/src/lib/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const transactionBuilder_1 = require("./transactionBuilder");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("./utils"));
const unsignedTransactionBuilder_1 = require("./unsignedTransactionBuilder");
const iface_1 = require("./iface");
const assert_1 = __importDefault(require("assert"));
class TransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        super.initBuilder(tx);
    }
    /** @inheritdoc */
    async buildImplementation() {
        // The ICP chain sets a memo field with a default value of 0. This ensures compatibility
        // by setting the memo to 0 if it's not explicitly provided.
        if (!this._memo || this._memo === undefined || this._memo === null) {
            this._memo = iface_1.DEFAULT_MEMO;
        }
        this.validateTransaction(this._transaction);
        this.buildIcpTransactionData();
        const unsignedTransactionBuilder = new unsignedTransactionBuilder_1.UnsignedTransactionBuilder(this._transaction.icpTransaction, this._coinConfig);
        const payloadsData = await unsignedTransactionBuilder.getUnsignedTransaction();
        this._transaction.payloadsData = payloadsData;
        return this._transaction;
    }
    buildIcpTransactionData() {
        (0, assert_1.default)(this._sender, new sdk_core_1.BuildTransactionError('sender is required before building'));
        (0, assert_1.default)(this._publicKey, new sdk_core_1.BuildTransactionError('sender public key is required before building'));
        (0, assert_1.default)(this._amount, new sdk_core_1.BuildTransactionError('amount is required before building'));
        (0, assert_1.default)(this._receiverId, new sdk_core_1.BuildTransactionError('receiver is required before building'));
        const publicKey = {
            hex_bytes: this._publicKey,
            curve_type: iface_1.CurveType.SECP256K1,
        };
        const senderOperation = {
            type: iface_1.OperationType.TRANSACTION,
            account: { address: this._sender },
            amount: {
                value: `-${this._amount}`,
                currency: {
                    symbol: this._coinConfig.family,
                    decimals: this._coinConfig.decimalPlaces,
                },
            },
        };
        const receiverOperation = {
            type: iface_1.OperationType.TRANSACTION,
            account: { address: this._receiverId },
            amount: {
                value: this._amount,
                currency: {
                    symbol: this._coinConfig.family,
                    decimals: this._coinConfig.decimalPlaces,
                },
            },
        };
        const feeOperation = {
            type: iface_1.OperationType.FEE,
            account: { address: this._sender },
            amount: {
                value: utils_1.default.feeData(),
                currency: {
                    symbol: this._coinConfig.family,
                    decimals: this._coinConfig.decimalPlaces,
                },
            },
        };
        const createdTimestamp = this._transaction.createdTimestamp;
        const { metaData, ingressEndTime } = utils_1.default.getMetaData(this._memo, createdTimestamp, this._ingressEnd);
        const icpTransaction = {
            public_keys: [publicKey],
            operations: [senderOperation, receiverOperation, feeOperation],
            metadata: metaData,
        };
        const icpTransactionData = {
            senderAddress: this._sender,
            receiverAddress: this._receiverId,
            amount: this._amount,
            fee: utils_1.default.feeData(),
            senderPublicKeyHex: this._publicKey,
            transactionType: iface_1.OperationType.TRANSACTION,
            expiryTime: ingressEndTime,
            memo: this._memo,
        };
        this._transaction.icpTransactionData = icpTransactionData;
        this._transaction.icpTransaction = icpTransaction;
    }
    /** @inheritdoc */
    signImplementation(key) {
        const signatures = utils_1.default.getSignatures(this._transaction.payloadsData, this._publicKey, key.key);
        this._transaction.addSignature(signatures);
        return this._transaction;
    }
}
exports.TransferBuilder = TransferBuilder;
