// @bitgo/sdk-coin-icp/dist/src/lib/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const iface_1 = require("./iface");
const signedTransactionBuilder_1 = require("./signedTransactionBuilder");
const transaction_1 = require("./transaction");
const utils_1 = __importDefault(require("./utils"));
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transaction_1.Transaction(_coinConfig);
    }
    signaturePayload() {
        return this._transaction.signaturePayload;
    }
    payloadData() {
        return this._transaction.payloadsData;
    }
    icpTransaction() {
        return this._transaction.icpTransaction;
    }
    /**
     * Sets the public key and the address of the sender of this transaction.
     *
     * @param {string} address the account that is sending this transaction
     * @param {string} pubKey the public key that is sending this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    sender(address, pubKey) {
        if (!address || !utils_1.default.isValidAddress(address.toString())) {
            throw new sdk_core_1.BuildTransactionError('Invalid or missing address, got: ' + address);
        }
        if (!pubKey || !utils_1.default.isValidPublicKey(pubKey)) {
            throw new sdk_core_1.BuildTransactionError('Invalid or missing pubKey, got: ' + pubKey);
        }
        this._sender = address;
        this._publicKey = pubKey;
        return this;
    }
    /**
     * Set the memo
     *
     * @param {number} memo - number that to be used as memo
     * @returns {TransactionBuilder} This transaction builder
     */
    memo(memo) {
        if (typeof memo !== 'number' || Number.isNaN(memo) || memo < 0) {
            throw new sdk_core_1.BuildTransactionError(`Invalid memo: ${memo}`);
        }
        this._memo = memo;
        return this;
    }
    /**
     * Set the ingressEnd timestamp
     * @param {number} ingressEnd - timestamp in nanoseconds
     * @returns {TransactionBuilder} This transaction builder
     */
    ingressEnd(ingressEnd) {
        if (BigInt(ingressEnd.toString()) < 0n) {
            throw new sdk_core_1.BuildTransactionError(`Invalid timestamp: ${ingressEnd}`);
        }
        this._ingressEnd = ingressEnd;
        return this;
    }
    /**
     * Sets the account Id of the receiver of this transaction.
     *
     * @param {string} accountId the account id of the account that is receiving this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    receiverId(accountId) {
        if (!accountId || !utils_1.default.isValidAddress(accountId)) {
            throw new sdk_core_1.BuildTransactionError('Invalid or missing accountId for receiver, got: ' + accountId);
        }
        this._receiverId = accountId;
        return this;
    }
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    /** @inheritdoc */
    set transaction(transaction) {
        this._transaction = transaction;
    }
    get transactionType() {
        return this._transaction.icpTransactionData.transactionType;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        this.transaction.fromRawTransaction(rawTransaction);
        const icpTransactionData = this.transaction.icpTransactionData;
        this.validateRawTransaction(icpTransactionData);
        this.buildImplementation();
        return this.transaction;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction || !transaction.icpTransactionData) {
            return;
        }
        utils_1.default.validateRawTransaction(transaction.icpTransactionData);
    }
    /**
     * Sets the amount of this transaction.
     *
     * @param {string} value the amount to be sent in e8s (1 ICP = 1e8 e8s)
     * @returns {TransactionBuilder} This transaction builder
     */
    amount(value) {
        utils_1.default.validateValue(new bignumber_js_1.default(value));
        this._amount = value;
        return this;
    }
    validateValue(value) {
        utils_1.default.validateValue(new bignumber_js_1.default(value));
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        this._transaction = tx;
        const icpTransactionData = tx.icpTransactionData;
        this._sender = icpTransactionData.senderAddress;
        this._receiverId = icpTransactionData.receiverAddress;
        this._publicKey = icpTransactionData.senderPublicKeyHex;
        this._amount = icpTransactionData.amount;
        this._memo = icpTransactionData.memo ?? iface_1.DEFAULT_MEMO;
        this._ingressEnd = Number(icpTransactionData.expiryTime);
    }
    validateAddress(address) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address');
        }
    }
    validateRawTransaction(rawTransaction) {
        utils_1.default.validateRawTransaction(rawTransaction);
    }
    /** @inheritdoc */
    validateKey(key) {
        if (!key || !key.key) {
            throw new sdk_core_1.SigningError('Key is required');
        }
        if (!utils_1.default.isValidPrivateKey(key.key)) {
            throw new sdk_core_1.SigningError('Invalid private key');
        }
    }
    /**
     * Combines the unsigned transaction and the signature payload to create a signed transaction.
     */
    combine() {
        const signedTransactionBuilder = new signedTransactionBuilder_1.SignedTransactionBuilder(this._transaction.unsignedTransaction, this._transaction.signaturePayload);
        this._transaction.signedTransaction = signedTransactionBuilder.getSignTransaction();
    }
}
exports.TransactionBuilder = TransactionBuilder;
