// @bitgo/sdk-coin-icp/dist/src/lib/transactionBuilderFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transaction_1 = require("./transaction");
const transferBuilder_1 = require("./transferBuilder");
const iface_1 = require("./iface");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    async from(rawTransaction) {
        const transaction = new transaction_1.Transaction(this._coinConfig);
        await transaction.fromRawTransaction(rawTransaction);
        try {
            switch (transaction.icpTransactionData.transactionType) {
                case iface_1.OperationType.TRANSACTION:
                    return this.getTransferBuilder(transaction);
                default:
                    throw new sdk_core_1.InvalidTransactionError('Invalid transaction');
            }
        }
        catch (e) {
            throw new sdk_core_1.InvalidTransactionError('Invalid transaction: ' + e.message);
        }
    }
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    static initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
    /** @inheritdoc */
    getTransferBuilder(tx) {
        return TransactionBuilderFactory.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    parseTransaction(rawTransaction, isSigned) {
        if (!rawTransaction) {
            throw new sdk_core_1.InvalidTransactionError('Transaction is empty');
        }
        const transaction = new transaction_1.Transaction(this._coinConfig);
        if (isSigned) {
            return transaction.parseSignedTransaction(rawTransaction);
        }
        return transaction.parseUnsignedTransaction(rawTransaction);
    }
    /** @inheritdoc */
    getWalletInitializationBuilder() {
        throw new sdk_core_1.MethodNotImplementedError();
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
