// @bitgo/sdk-coin-icp/dist/src/lib/unsignedTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsignedTransactionBuilder = void 0;
const iface_1 = require("./iface");
const utils_1 = __importDefault(require("./utils"));
const statics_1 = require("@bitgo/statics");
class UnsignedTransactionBuilder {
    constructor(icpTransactionPayload, staticsCoin) {
        this._icpTransactionPayload = icpTransactionPayload;
        this._staticsCoin = staticsCoin;
    }
    async getUnsignedTransaction() {
        // interval value can be altered here to change the count of ingressExpiries which eventually changes the number of interval value can be altered here to change the count of ingressExpiries which eventually changes the number of payloads that needs to be signed
        const interval = Number(this._icpTransactionPayload.metadata.ingress_end) -
            Number(this._icpTransactionPayload.metadata.ingress_start); // 300s in nanoseconds
        const ingressExpiries = this.getIngressExpiries(this._icpTransactionPayload.metadata.ingress_start, this._icpTransactionPayload.metadata.ingress_end, interval);
        const sendArgs = this.getSendArgs(this._icpTransactionPayload.metadata.memo, this._icpTransactionPayload.metadata.created_at_time, this._icpTransactionPayload.operations[1].amount.value, this._icpTransactionPayload.operations[2].amount.value, this._icpTransactionPayload.operations[1].account.address);
        const update = await this.getUpdate(sendArgs, this._icpTransactionPayload.public_keys[0].hex_bytes);
        const updates = [];
        updates.push([iface_1.OperationType.TRANSACTION, update]);
        const txn = { updates: updates, ingress_expiries: ingressExpiries };
        const unsignedTransaction = utils_1.default.cborEncode(txn);
        const payloads = [];
        this.getPayloads(payloads, ingressExpiries, this._icpTransactionPayload.operations[0].account.address, update);
        const payloadsData = {
            payloads: payloads,
            unsigned_transaction: unsignedTransaction,
        };
        return payloadsData;
    }
    getPayloads(payloads, ingressExpiries, accountAddress, update) {
        if (ingressExpiries.length != 1) {
            throw new Error('ingress expiry can have only one entry');
        }
        const ingressExpiry = ingressExpiries[0];
        const clonedUpdate = {
            canister_id: Buffer.from(update.canister_id),
            method_name: update.method_name,
            arg: update.arg,
            sender: update.sender,
            ingress_expiry: ingressExpiry,
        };
        const representationIndependentHash = utils_1.default.HttpCanisterUpdateRepresentationIndependentHash(clonedUpdate);
        const transactionPayload = {
            hex_bytes: utils_1.default.blobToHex(utils_1.default.makeSignatureData(representationIndependentHash)),
            account_identifier: { address: accountAddress },
            signature_type: iface_1.SignatureType.ECDSA,
        };
        payloads.push(transactionPayload);
        return payloads;
    }
    getIngressExpiries(ingressStartTime, ingressEndTime, interval) {
        const ingressExpiries = [];
        for (let now = Number(ingressStartTime); now < Number(ingressEndTime); now += interval) {
            const ingressExpiry = BigInt(now + (iface_1.MAX_INGRESS_TTL - iface_1.PERMITTED_DRIFT));
            ingressExpiries.push(ingressExpiry);
        }
        return ingressExpiries;
    }
    getSendArgs(memo, created_at_time, amount, fee, receiver) {
        const sendArgs = {
            payment: { receiverGets: { e8s: Number(amount) } },
            maxFee: { e8s: -Number(fee) },
            to: { hash: Buffer.from(receiver, 'hex') },
            createdAtTime: { timestampNanos: Number(created_at_time) },
            memo: { memo: memo },
        };
        return sendArgs;
    }
    async getUpdate(sendArgs, publicKeyHex) {
        const principalId = utils_1.default.getPrincipalIdFromPublicKey(publicKeyHex).toUint8Array();
        const senderBlob = Buffer.from(principalId);
        const ledgerCanisterId = this._staticsCoin.network.type === statics_1.NetworkType.TESTNET ? iface_1.TESTNET_LEDGER_CANISTER_ID : iface_1.LEDGER_CANISTER_ID;
        const canisterIdBuffer = Buffer.from(ledgerCanisterId);
        const args = await utils_1.default.toArg(sendArgs);
        const update = {
            canister_id: canisterIdBuffer,
            method_name: iface_1.MethodName.SEND_PB,
            arg: args,
            sender: senderBlob,
            ingress_expiry: 0n,
        };
        return update;
    }
}
exports.UnsignedTransactionBuilder = UnsignedTransactionBuilder;
