// @bitgo/abstract-utxo/dist/cjs/src/keychains.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeySignatures = exports.UtxoNamedKeychains = exports.UtxoKeychain = void 0;
exports.toKeychainTriple = toKeychainTriple;
exports.toBip32Triple = toBip32Triple;
exports.toXpubTriple = toXpubTriple;
exports.fetchKeychains = fetchKeychains;
exports.fetchWasmRootWalletKeys = fetchWasmRootWalletKeys;
exports.getKeySignatures = getKeySignatures;
const assert_1 = __importDefault(require("assert"));
const t = __importStar(require("io-ts"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const sdk_core_1 = require("@bitgo/sdk-core");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
/*

The standard Keychain type from sdk-core requires a bunch of uninteresting parameters like `id` and `type` and leaves
important fields like `pub` and `prv` optional.

This is a more focused type that only includes the fields we care about.
 */
/**
 * A keychain for a UTXO wallet.
 */
exports.UtxoKeychain = t.intersection([
    t.type({
        pub: t.string,
    }),
    t.partial({
        prv: t.string,
        encryptedPrv: t.string,
    }),
], 'UtxoKeychain');
exports.UtxoNamedKeychains = t.type({
    user: exports.UtxoKeychain,
    backup: exports.UtxoKeychain,
    bitgo: exports.UtxoKeychain,
});
function toKeychainTriple(keychains) {
    const { user, backup, bitgo } = keychains;
    return [user, backup, bitgo];
}
function toBip32Triple(keychains) {
    if (keychains instanceof utxo_lib_1.bitgo.RootWalletKeys) {
        return keychains.triple.map((k) => wasm_utxo_1.BIP32.fromBase58(k.toBase58()));
    }
    if (Array.isArray(keychains)) {
        if (keychains.length !== 3) {
            throw new Error('expected 3 keychains');
        }
        return keychains.map((keychain) => {
            const v = typeof keychain === 'string' ? keychain : keychain.pub;
            return wasm_utxo_1.BIP32.fromBase58(v);
        });
    }
    return toBip32Triple(toKeychainTriple(keychains));
}
function toXpub(keychain) {
    if (typeof keychain === 'string') {
        if (keychain.startsWith('xpub')) {
            return keychain;
        }
        throw new Error('expected xpub');
    }
    if ('neutered' in keychain) {
        return keychain.neutered().toBase58();
    }
    if ('pub' in keychain) {
        return toXpub(keychain.pub);
    }
    throw new Error('expected keychain');
}
function toXpubTriple(keychains) {
    if (Array.isArray(keychains)) {
        if (keychains.length !== 3) {
            throw new Error('expected 3 keychains');
        }
        return keychains.map((k) => toXpub(k));
    }
    return toXpubTriple(toKeychainTriple(keychains));
}
async function fetchKeychains(coin, wallet, reqId) {
    const result = await (0, sdk_core_1.promiseProps)({
        user: coin.keychains().get({ id: wallet.keyIds()[sdk_core_1.KeyIndices.USER], reqId }),
        backup: coin.keychains().get({ id: wallet.keyIds()[sdk_core_1.KeyIndices.BACKUP], reqId }),
        bitgo: coin.keychains().get({ id: wallet.keyIds()[sdk_core_1.KeyIndices.BITGO], reqId }),
    });
    (0, assert_1.default)(exports.UtxoNamedKeychains.is(result));
    return result;
}
/**
 * Fetch wallet keys as wasm-utxo RootWalletKeys
 */
async function fetchWasmRootWalletKeys(coin, wallet, reqId) {
    const keychains = await fetchKeychains(coin, wallet, reqId);
    return wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from([keychains.user.pub, keychains.backup.pub, keychains.bitgo.pub]);
}
exports.KeySignatures = t.partial({
    backupPub: t.string,
    bitgoPub: t.string,
});
function getKeySignatures(wallet) {
    if (t.partial({ keySignatures: exports.KeySignatures }).is(wallet._wallet)) {
        return wallet._wallet.keySignatures;
    }
    return undefined;
}
