// @bitgo/abstract-utxo/dist/cjs/src/names.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNetworkFromChain = exports.utxoCoinsTestnet = exports.utxoCoinsMainnet = void 0;
exports.isUtxoCoinNameMainnet = isUtxoCoinNameMainnet;
exports.isUtxoCoinNameTestnet = isUtxoCoinNameTestnet;
exports.isUtxoCoinName = isUtxoCoinName;
exports.getMainnetCoinName = getMainnetCoinName;
exports.getFamilyFromNetwork = getFamilyFromNetwork;
exports.getCoinName = getCoinName;
exports.getNetworkFromCoinName = getNetworkFromCoinName;
exports.getFullNameFromCoinName = getFullNameFromCoinName;
exports.getFullNameFromNetwork = getFullNameFromNetwork;
exports.isTestnetCoin = isTestnetCoin;
exports.isMainnetCoin = isMainnetCoin;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
exports.utxoCoinsMainnet = ['btc', 'bch', 'bcha', 'bsv', 'btg', 'dash', 'doge', 'ltc', 'zec'];
exports.utxoCoinsTestnet = [
    'tbtc',
    'tbtc4',
    'tbtcsig',
    'tbtcbgsig',
    'tbch',
    'tbcha',
    'tbsv',
    'tbtg',
    'tdash',
    'tdoge',
    'tltc',
    'tzec',
];
function isUtxoCoinNameMainnet(coinName) {
    return exports.utxoCoinsMainnet.includes(coinName);
}
function isUtxoCoinNameTestnet(coinName) {
    return exports.utxoCoinsTestnet.includes(coinName);
}
function isUtxoCoinName(coinName) {
    return isUtxoCoinNameMainnet(coinName) || isUtxoCoinNameTestnet(coinName);
}
function getMainnetCoinName(coinName) {
    if (isUtxoCoinNameMainnet(coinName)) {
        return coinName;
    }
    switch (coinName) {
        case 'tbtc4':
        case 'tbtcsig':
        case 'tbtcbgsig':
            return 'btc';
        default:
            return coinName.slice(1);
    }
}
function getNetworkName(n) {
    const name = utxolib.getNetworkName(n);
    if (!name) {
        throw new Error('Unknown network');
    }
    return name;
}
/**
 * @deprecated - will be removed when we drop support for utxolib
 * @param n
 * @returns the family name for a network. Testnets and mainnets of the same coin share the same family name.
 */
function getFamilyFromNetwork(n) {
    switch (getNetworkName(n)) {
        case 'bitcoin':
        case 'testnet':
        case 'bitcoinPublicSignet':
        case 'bitcoinTestnet4':
        case 'bitcoinBitGoSignet':
            return 'btc';
        case 'bitcoincash':
        case 'bitcoincashTestnet':
            return 'bch';
        case 'ecash':
        case 'ecashTest':
            return 'bcha';
        case 'bitcoingold':
        case 'bitcoingoldTestnet':
            return 'btg';
        case 'bitcoinsv':
        case 'bitcoinsvTestnet':
            return 'bsv';
        case 'dash':
        case 'dashTest':
            return 'dash';
        case 'dogecoin':
        case 'dogecoinTest':
            return 'doge';
        case 'litecoin':
        case 'litecoinTest':
            return 'ltc';
        case 'zcash':
        case 'zcashTest':
            return 'zec';
    }
}
/**
 * @deprecated - will be removed when we drop support for utxolib
 * Get the chain name for a network.
 * The chain is different for every network.
 */
function getCoinName(n) {
    switch (getNetworkName(n)) {
        case 'bitcoinPublicSignet':
            return 'tbtcsig';
        case 'bitcoinTestnet4':
            return 'tbtc4';
        case 'bitcoinBitGoSignet':
            return 'tbtcbgsig';
        case 'bitcoin':
        case 'testnet':
        case 'bitcoincash':
        case 'bitcoincashTestnet':
        case 'ecash':
        case 'ecashTest':
        case 'bitcoingold':
        case 'bitcoingoldTestnet':
        case 'bitcoinsv':
        case 'bitcoinsvTestnet':
        case 'dash':
        case 'dashTest':
        case 'dogecoin':
        case 'dogecoinTest':
        case 'litecoin':
        case 'litecoinTest':
        case 'zcash':
        case 'zcashTest':
            const mainnetName = getFamilyFromNetwork(n);
            return utxolib.isTestnet(n) ? `t${mainnetName}` : mainnetName;
    }
}
/**
 * @deprecated - will be removed when we drop support for utxolib
 * @param coinName - the name of the coin (e.g. 'btc', 'bch', 'ltc'). Also called 'chain' in some contexts.
 * @returns the network for a coin. This is the mainnet network for the coin.
 */
function getNetworkFromCoinName(coinName) {
    for (const network of utxolib.getNetworkList()) {
        if (getCoinName(network) === coinName) {
            return network;
        }
    }
    throw new Error(`Unknown coin name ${coinName}`);
}
/** @deprecated - use getNetworkFromCoinName instead */
exports.getNetworkFromChain = getNetworkFromCoinName;
function getBaseNameFromMainnet(coinName) {
    switch (coinName) {
        case 'btc':
            return 'Bitcoin';
        case 'bch':
            return 'Bitcoin Cash';
        case 'bcha':
            return 'Bitcoin ABC';
        case 'btg':
            return 'Bitcoin Gold';
        case 'bsv':
            return 'Bitcoin SV';
        case 'dash':
            return 'Dash';
        case 'doge':
            return 'Dogecoin';
        case 'ltc':
            return 'Litecoin';
        case 'zec':
            return 'ZCash';
    }
}
function getFullNameFromCoinName(coinName) {
    let prefix;
    switch (coinName) {
        case 'tbtc4':
            prefix = 'Testnet4 ';
            break;
        case 'tbtcsig':
            prefix = 'Public Signet ';
            break;
        case 'tbtcbgsig':
            prefix = 'BitGo Signet ';
            break;
        default:
            prefix = isUtxoCoinNameTestnet(coinName) ? 'Testnet ' : '';
    }
    return prefix + getBaseNameFromMainnet(getMainnetCoinName(coinName));
}
/** @deprecated - use getFullNameFromCoinName instead */
function getFullNameFromNetwork(n) {
    return getFullNameFromCoinName(getCoinName(n));
}
function isTestnetCoin(coinName) {
    return isUtxoCoinNameTestnet(coinName);
}
function isMainnetCoin(coinName) {
    return isUtxoCoinNameMainnet(coinName);
}
