// @bitgo/abstract-utxo/dist/cjs/src/wasmUtil.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.toWasmBIP32 = toWasmBIP32;
exports.toUtxolibBIP32 = toUtxolibBIP32;
exports.toWasmECPair = toWasmECPair;
exports.isUtxoLibPsbt = isUtxoLibPsbt;
exports.toWasmPsbt = toWasmPsbt;
exports.sumValues = sumValues;
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const utxolib = __importStar(require("@bitgo/utxo-lib"));
/**
 * Convert a utxo-lib BIP32Interface to a wasm-utxo BIP32 instance.
 * Preserves private key by using base58 serialization.
 */
function toWasmBIP32(key) {
    if (key instanceof wasm_utxo_1.BIP32) {
        return key;
    }
    // All utxo-lib BIP32Interface instances have toBase58
    return wasm_utxo_1.BIP32.fromBase58(key.toBase58());
}
/**
 * Convert a wasm-utxo BIP32 to a utxo-lib BIP32Interface.
 * Used at boundaries where utxo-lib APIs require their own BIP32Interface type.
 */
function toUtxolibBIP32(key) {
    return utxolib.bip32.fromBase58(key.toBase58());
}
function toWasmECPair(key) {
    if (key instanceof wasm_utxo_1.ECPair) {
        return key;
    }
    if (key instanceof Uint8Array) {
        return wasm_utxo_1.ECPair.from(key);
    }
    if (key.privateKey) {
        return wasm_utxo_1.ECPair.fromPrivateKey(key.privateKey);
    }
    return wasm_utxo_1.ECPair.fromPublicKey(key.publicKey);
}
function isUtxoLibPsbt(psbt) {
    return psbt instanceof utxolib.Psbt || psbt instanceof utxolib.bitgo.UtxoPsbt;
}
function toWasmPsbt(psbt) {
    if (psbt instanceof wasm_utxo_1.Psbt) {
        return psbt;
    }
    if (psbt instanceof Uint8Array) {
        return wasm_utxo_1.Psbt.deserialize(psbt);
    }
    if (isUtxoLibPsbt(psbt)) {
        return wasm_utxo_1.Psbt.deserialize(psbt.toBuffer());
    }
    throw new Error('Unsupported PSBT type');
}
/**
 * Sum the `value` property of an array of objects.
 */
function sumValues(arr) {
    return arr.reduce((sum, e) => sum + e.value, BigInt(0));
}
