// @bitgo/utxo-core/dist/cjs/src/dustThreshold.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDustThresholdSat = getDustThresholdSat;
exports.getDustThresholdSatForOutputScript = getDustThresholdSatForOutputScript;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
/**
 * Checks if the output script is a witness script or not
 * @param script
 * @returns true if the script is a witness script
 */
function isWitnessOutputScript(script) {
    /**
     * Source: https://github.com/bitcoin/bitcoin/blob/v28.1/src/script/script.cpp#L241-L257
     * A witness program is any valid CScript that consists of a 1-byte push opcode
     * followed by a data push between 2 and 40 bytes.
     */
    if (script.length < 4 || script.length > 42) {
        return false;
    }
    if (script[0] !== utxolib.opcodes.OP_0 && (script[0] < utxolib.opcodes.OP_1 || script[0] > utxolib.opcodes.OP_16)) {
        return false;
    }
    return script[1] + 2 === script.length;
}
function getDustRelayLimit(network) {
    network = utxolib.getMainnet(network);
    switch (network) {
        case utxolib.networks.bitcoin:
        case utxolib.networks.bitcoingold:
        case utxolib.networks.dash:
            // btc:  https://github.com/bitcoin/bitcoin/blob/v28.0/src/policy/policy.h#L50-L55
            // btg:  https://github.com/BTCGPU/BTCGPU/blob/v0.17.3/src/policy/policy.h#L48
            // dash: https://github.com/dashpay/dash/blob/v22.0.0-beta.1/src/policy/policy.h#L41-L46
            return { feeRateSatKB: 3000 };
        case utxolib.networks.bitcoincash:
            // https://github.com/bitcoin-cash-node/bitcoin-cash-node/blob/v27.1.0/src/policy/policy.h#L76-L83
            // I actually haven't looked at BSV and am depressed that I still need to handle the case here
            return { feeRateSatKB: 1000 };
        case utxolib.networks.dogecoin:
            // https://github.com/dogecoin/dogecoin/blob/v1.14.8/src/policy/policy.h#L65-L81
            // (COIN / 100) / 10;
            return { satAmount: 1000000 };
        case utxolib.networks.litecoin:
            //  https://github.com/litecoin-project/litecoin/blob/master/src/policy/policy.h#L47-L52
            return { feeRateSatKB: 30000 };
        case utxolib.networks.zcash:
            // https://github.com/zcash/zcash/blob/master/src/primitives/transaction.h#L396-L399
            // https://github.com/zcash/zcash/blob/v6.0.0/src/policy/policy.h#L43-L89 (I don't quite get it)
            return { satAmount: 300 };
        case utxolib.networks.bitcoinsv:
            throw new Error('deprecated coin');
        default:
            throw new Error('unsupported network');
    }
}
function getSpendSize(network, outputSize, isWitness) {
    network = utxolib.getMainnet(network);
    switch (network) {
        case utxolib.networks.bitcoin:
        case utxolib.networks.bitcoincash:
        case utxolib.networks.bitcoingold:
        case utxolib.networks.litecoin:
            /*
              btc:  https://github.com/bitcoin/bitcoin/blob/v28.0/src/policy/policy.cpp#L26-L68
              bch:  https://github.com/bitcoin-cash-node/bitcoin-cash-node/blob/v27.1.0/src/policy/policy.cpp#L18-L36 (btc-ish)
              btg:  https://github.com/BTCGPU/BTCGPU/blob/v0.17.3/src/policy/policy.cpp#L18-L50 (btc-ish)
              ltc:  https://github.com/litecoin-project/litecoin/blob/v0.21.4/src/policy/policy.cpp#L15-L47 (btc-ish)
      
              The fixed component here is 69.75 for isWitness=true and 150 for isWitness=false.
             */
            return outputSize + 32 + 4 + 1 + 107 / (isWitness ? 4 : 1) + 4;
        case utxolib.networks.dash:
            // dash: https://github.com/dashpay/dash/blob/v21.1.1/src/policy/policy.cpp#L14-L30 (btc-ish)
            // how did they end up with 148? I don't know
            return outputSize + 148;
        case utxolib.networks.dogecoin:
        case utxolib.networks.zcash:
            // doge: https://github.com/dogecoin/dogecoin/blob/v1.14.8/src/policy/policy.h#L65-L81 (hardcoded)
            // zec:  https://github.com/zcash/zcash/blob/v6.0.0/src/policy/policy.h#L43-L89 (some weird other thing, doge-ish)
            throw new Error('dust limit is size-independent');
        case utxolib.networks.bitcoinsv:
            throw new Error('deprecated coin');
        default:
            throw new Error('unsupported network');
    }
}
function getDustThresholdSat(network, outputSize, isWitness) {
    const dustLimit = getDustRelayLimit(network);
    if ('satAmount' in dustLimit) {
        return dustLimit.satAmount;
    }
    if ('feeRateSatKB' in dustLimit) {
        const spendSize = getSpendSize(network, outputSize, isWitness);
        return Math.ceil((dustLimit.feeRateSatKB * spendSize) / 1000);
    }
    throw new Error('unexpected dustLimit');
}
function getDustThresholdSatForOutputScript(network, script) {
    return getDustThresholdSat(network, script.length, isWitnessOutputScript(script));
}
