// @bitgo/sdk-coin-ada/dist/src/lib/stakingPledgeBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakingPledgeBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
class StakingPledgeBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._type = sdk_core_1.TransactionType.StakingPledge;
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingPledge;
    }
    /** @inheritdoc */
    initBuilder(tx) {
        super.initBuilder(tx);
    }
    /**
     * Add node key witness to the transaction to the transaction's witnesses list.
     *
     * @param nodePublicKey Public node key provided by the user.
     * @param nodeKeySignature Signature of the node key provided by the user.
     */
    addNodeKeyWitness(nodePublicKey, nodeKeySignature) {
        this.addSignature({ pub: nodePublicKey }, Buffer.from(nodeKeySignature, 'hex'));
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        const tx = await super.buildImplementation();
        tx.setTransactionType(sdk_core_1.TransactionType.StakingPledge);
        return tx;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        return super.fromImplementation(rawTransaction);
    }
    signImplementation(key) {
        return super.signImplementation(key);
    }
}
exports.StakingPledgeBuilder = StakingPledgeBuilder;
