// @vechain/sdk-errors/dist/index.js
'use strict';

var rr=Object.defineProperty;var tr=(t,a)=>{for(var o in a)rr(t,o,{get:a[o],enumerable:!0});};var I={};tr(I,{CannotFindTransaction:()=>R,CertificateSignatureMismatch:()=>y,ContractCallError:()=>b,ContractDeploymentFailed:()=>u,InvalidAbiDataToEncodeOrDecode:()=>f,InvalidAbiItem:()=>i,InvalidAbiSignatureFormat:()=>E,InvalidAddress:()=>g,InvalidBloom:()=>d,InvalidBloomParams:()=>D,InvalidCommandLineArguments:()=>A,InvalidConfigurationFile:()=>L,InvalidConfigurationFilePath:()=>q,InvalidDataType:()=>j,InvalidHDKey:()=>w,InvalidHDKeyMnemonic:()=>O,InvalidHTTPParams:()=>V,InvalidHTTPRequest:()=>h,InvalidKeystore:()=>k,InvalidKeystoreParams:()=>K,InvalidOperation:()=>P,InvalidRLP:()=>_,InvalidSecp256k1MessageHash:()=>B,InvalidSecp256k1PrivateKey:()=>T,InvalidSecp256k1Signature:()=>H,InvalidTransactionField:()=>N,InvalidTransactionType:()=>C,JSONRPCInternalError:()=>Q,JSONRPCInvalidDefaultBlock:()=>X,JSONRPCInvalidParams:()=>G,JSONRPCInvalidRequest:()=>z,JSONRPCMethodNotFound:()=>W,JSONRPCMethodNotImplemented:()=>Z,JSONRPCParseError:()=>U,JSONRPCProviderError:()=>c,JSONRPCServerError:()=>Y,NotDelegatedTransaction:()=>F,PollExecution:()=>S,ProviderMethodError:()=>$,SignerMethodError:()=>M,UnavailableTransactionField:()=>J,UnsupportedOperation:()=>l,VechainSDKError:()=>r,assertInnerError:()=>x,createErrorMessage:()=>m,stringifyData:()=>p});var p=t=>JSON.stringify(t,(()=>{let o=new WeakSet;return (s,e)=>{if(typeof e=="object"&&e!==null&&s!==""){if(o.has(e))return;o.add(e);}return typeof e=="bigint"?e.toString()+"n":e}})(),2);function x(t){return t instanceof Error?t:new Error(`Inner error is not an instance of Error. Object:
	${p(t)}`)}function m(t,a,o,s){let e=s?.message!==void 0?`
-Internal error: 
	${s.message}`:"";return `Method '${t}' failed.
-Reason: '${a}'
-Parameters: 
	${p(o)}`+e}var r=class extends Error{constructor(o,s,e,n){super(m(o,s,e,n===void 0?void 0:x(n)));this.methodName=o;this.errorMessage=s;this.data=e;this.innerError=n;}};var f=class extends r{},i=class extends r{},E=class extends r{};var g=class extends r{};var d=class extends r{},D=class extends r{};var y=class extends r{};var u=class extends r{},b=class extends r{};var j=class extends r{},l=class extends r{};var O=class extends r{},w=class extends r{};var h=class extends r{},V=class extends r{};var k=class extends r{},K=class extends r{};var S=class extends r{};var $=class extends r{},c=class extends r{constructor(o,s,e,n,v){super(o,e,{code:s,message:e,data:n},v);this.methodName=o;this.innerError=v;}},U=class extends c{constructor(o,s,e,n){super(o,-32700,s,e,n);this.methodName=o;this.innerError=n;}},z=class extends c{constructor(o,s,e,n){super(o,-32600,s,e,n);this.methodName=o;this.innerError=n;}},W=class extends c{constructor(o,s,e,n){super(o,-32601,s,e,n);this.methodName=o;this.innerError=n;}},G=class extends c{constructor(o,s,e,n){super(o,-32602,s,e,n);this.methodName=o;this.innerError=n;}},Q=class extends c{constructor(o,s,e,n){super(o,-32603,s,e,n);this.methodName=o;this.innerError=n;}},X=class extends r{},Y=class extends c{constructor(o,s,e,n){super(o,-32e3,s,e,n);this.methodName=o;this.innerError=n;}},Z=class extends c{constructor(o,s,e,n){super(o,-32004,s,e,n);this.methodName=o;this.innerError=n;}};var _=class extends r{};var A=class extends r{},q=class extends r{},L=class extends r{};var T=class extends r{},B=class extends r{},H=class extends r{};var M=class extends r{};var J=class extends r{},N=class extends r{},F=class extends r{},R=class extends r{},C=class extends r{};var P=class extends r{};

exports.CannotFindTransaction = R;
exports.CertificateSignatureMismatch = y;
exports.ContractCallError = b;
exports.ContractDeploymentFailed = u;
exports.InvalidAbiDataToEncodeOrDecode = f;
exports.InvalidAbiItem = i;
exports.InvalidAbiSignatureFormat = E;
exports.InvalidAddress = g;
exports.InvalidBloom = d;
exports.InvalidBloomParams = D;
exports.InvalidCommandLineArguments = A;
exports.InvalidConfigurationFile = L;
exports.InvalidConfigurationFilePath = q;
exports.InvalidDataType = j;
exports.InvalidHDKey = w;
exports.InvalidHDKeyMnemonic = O;
exports.InvalidHTTPParams = V;
exports.InvalidHTTPRequest = h;
exports.InvalidKeystore = k;
exports.InvalidKeystoreParams = K;
exports.InvalidOperation = P;
exports.InvalidRLP = _;
exports.InvalidSecp256k1MessageHash = B;
exports.InvalidSecp256k1PrivateKey = T;
exports.InvalidSecp256k1Signature = H;
exports.InvalidTransactionField = N;
exports.InvalidTransactionType = C;
exports.JSONRPCInternalError = Q;
exports.JSONRPCInvalidDefaultBlock = X;
exports.JSONRPCInvalidParams = G;
exports.JSONRPCInvalidRequest = z;
exports.JSONRPCMethodNotFound = W;
exports.JSONRPCMethodNotImplemented = Z;
exports.JSONRPCParseError = U;
exports.JSONRPCProviderError = c;
exports.JSONRPCServerError = Y;
exports.NotDelegatedTransaction = F;
exports.PollExecution = S;
exports.ProviderMethodError = $;
exports.SignerMethodError = M;
exports.UnavailableTransactionField = J;
exports.UnsupportedOperation = l;
exports.VechainSDKError = r;
exports.assertInnerError = x;
exports.createErrorMessage = m;
exports.errors = I;
exports.stringifyData = p;
