// @bitgo/wasm-utxo/dist/esm/js/fixedScriptWallet/ReplayProtection.js
import { WasmReplayProtection } from "../wasm/wasm_utxo.js";
import { ECPair } from "../ecpair.js";
/**
 * ReplayProtection wrapper class for PSBT replay protection inputs
 */
export class ReplayProtection {
    _wasm;
    constructor(_wasm) {
        this._wasm = _wasm;
    }
    /**
     * Create a ReplayProtection instance from a WasmReplayProtection instance (internal use)
     * @internal
     */
    static fromWasm(wasm) {
        return new ReplayProtection(wasm);
    }
    /**
     * Convert ReplayProtectionArg to ReplayProtection instance
     * @param arg - The replay protection in various formats
     * @param network - Optional network string (required for addresses variant)
     * @returns ReplayProtection instance
     */
    static from(arg, network) {
        // Short-circuit if already a ReplayProtection instance
        if (arg instanceof ReplayProtection) {
            return arg;
        }
        // If it's a WasmReplayProtection instance, wrap it
        if (arg instanceof WasmReplayProtection) {
            return new ReplayProtection(arg);
        }
        // Handle object variants
        if ("publicKeys" in arg) {
            // Convert ECPairArg to public key bytes
            const publicKeyBytes = arg.publicKeys.map((key) => ECPair.from(key).publicKey);
            const wasm = WasmReplayProtection.from_public_keys(publicKeyBytes);
            return new ReplayProtection(wasm);
        }
        if ("outputScripts" in arg) {
            const wasm = WasmReplayProtection.from_output_scripts(arg.outputScripts);
            return new ReplayProtection(wasm);
        }
        if ("addresses" in arg) {
            if (!network) {
                throw new Error("Network is required when using addresses variant");
            }
            const wasm = WasmReplayProtection.from_addresses(arg.addresses, network);
            return new ReplayProtection(wasm);
        }
        throw new Error("Invalid ReplayProtectionArg type");
    }
    /**
     * Create from public keys (derives P2SH-P2PK output scripts)
     * @param publicKeys - Array of ECPair instances or arguments
     * @returns ReplayProtection instance
     */
    static fromPublicKeys(publicKeys) {
        return ReplayProtection.from({ publicKeys });
    }
    /**
     * Create from output scripts
     * @param outputScripts - Array of output script buffers
     * @returns ReplayProtection instance
     */
    static fromOutputScripts(outputScripts) {
        return ReplayProtection.from({ outputScripts });
    }
    /**
     * Create from addresses
     * @param addresses - Array of address strings
     * @param network - Network string (e.g., "bitcoin", "testnet", "btc", "tbtc")
     * @returns ReplayProtection instance
     */
    static fromAddresses(addresses, network) {
        return ReplayProtection.from({ addresses }, network);
    }
    /**
     * Get the underlying WASM instance (internal use only)
     * @internal
     */
    get wasm() {
        return this._wasm;
    }
}
