// @bitgo/sdk-coin-atom/dist/test/unit/atom.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const sdk_api_1 = require("@bitgo/sdk-api");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const sdk_test_1 = require("@bitgo/sdk-test");
const statics_1 = require("@bitgo/statics");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sinon_1 = __importDefault(require("sinon"));
const src_1 = require("../../src");
const constants_1 = require("../../src/lib/constants");
const utils_1 = __importDefault(require("../../src/lib/utils"));
const atom_1 = require("../resources/atom");
const should = require("should");
describe('ATOM', function () {
    let bitgo;
    let basecoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister('atom', src_1.Atom.createInstance);
        bitgo.safeRegister('tatom', src_1.Tatom.createInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin('tatom');
    });
    it('should retun the right info', function () {
        const atom = bitgo.coin('atom');
        const tatom = bitgo.coin('tatom');
        atom.getChain().should.equal('atom');
        atom.getFamily().should.equal('atom');
        atom.getFullName().should.equal('Cosmos Hub ATOM');
        atom.getBaseFactor().should.equal(1e6);
        tatom.getChain().should.equal('tatom');
        tatom.getFamily().should.equal('atom');
        tatom.getFullName().should.equal('Testnet Cosmos Hub ATOM');
        tatom.getBaseFactor().should.equal(1e6);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const addressDetails = basecoin.getAddressDetails(atom_1.address.noMemoIdAddress);
            addressDetails.address.should.equal(atom_1.address.noMemoIdAddress);
            should.not.exist(addressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const addressDetails = basecoin.getAddressDetails(atom_1.address.validMemoIdAddress);
            addressDetails.address.should.equal(atom_1.address.validMemoIdAddress.split('?')[0]);
            addressDetails.memoId.should.equal('2');
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                basecoin.getAddressDetails(atom_1.address.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const receiveAddress = {
                address: 'cosmos1r7k3xmlljsfdk6y9r8ssfnt4ks5ka0nwmm0z70?memoId=7',
                coinSpecific: {
                    rootAddress: 'cosmos1r7k3xmlljsfdk6y9r8ssfnt4ks5ka0nwmm0z70',
                    memoID: '7',
                },
            };
            const isValid = await basecoin.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(utils_1.default.isValidAddress(atom_1.address.address1), true);
            should.equal(utils_1.default.isValidAddress(atom_1.address.address2), true);
            should.equal(utils_1.default.isValidAddress(atom_1.address.address3), false);
            should.equal(utils_1.default.isValidAddress(atom_1.address.address4), false);
            should.equal(utils_1.default.isValidAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidAddress(undefined), false);
            should.equal(utils_1.default.isValidAddress(''), false);
            should.equal(utils_1.default.isValidAddress(atom_1.address.validMemoIdAddress), true);
            should.equal(utils_1.default.isValidAddress(atom_1.address.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(utils_1.default.isValidValidatorAddress(atom_1.address.validatorAddress1), true);
            should.equal(utils_1.default.isValidValidatorAddress(atom_1.address.validatorAddress2), true);
            should.equal(utils_1.default.isValidValidatorAddress(atom_1.address.validatorAddress3), false);
            should.equal(utils_1.default.isValidValidatorAddress(atom_1.address.validatorAddress4), false);
            should.equal(utils_1.default.isValidValidatorAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidValidatorAddress(undefined), false);
            should.equal(utils_1.default.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: atom_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: atom_1.TEST_SEND_TX.recipient,
                        amount: atom_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify sendMany transaction', async function () {
            const txPrebuild = {
                txHex: atom_1.TEST_SEND_MANY_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: 'cosmos16ghn9c6f5yua09zqw7y794mvc30h4y4md7ckuk',
                        amount: '500',
                    },
                    {
                        address: 'cosmos1ytez06yx0u3yjzjjjm02xyx3mh25akenzql3n8',
                        amount: '500',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify delegate transaction', async function () {
            const txPrebuild = {
                txHex: atom_1.TEST_DELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: atom_1.TEST_DELEGATE_TX.validator,
                        amount: atom_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify undelegate transaction', async function () {
            const txPrebuild = {
                txHex: atom_1.TEST_UNDELEGATE_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: atom_1.TEST_UNDELEGATE_TX.validator,
                        amount: atom_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify withdraw rewards transaction', async function () {
            const txPrebuild = {
                txHex: atom_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: atom_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: atom_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: atom_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: atom_1.TEST_SEND_TX.recipient,
                        amount: atom_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: atom_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: atom_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain sendMany transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: atom_1.TEST_SEND_MANY_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: atom_1.TEST_SEND_MANY_TX.hash,
                outputs: [
                    {
                        address: 'cosmos16ghn9c6f5yua09zqw7y794mvc30h4y4md7ckuk',
                        amount: '500',
                    },
                    {
                        address: 'cosmos1ytez06yx0u3yjzjjjm02xyx3mh25akenzql3n8',
                        amount: '500',
                    },
                ],
                outputAmount: '1000',
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: atom_1.TEST_SEND_MANY_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a delegate transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: atom_1.TEST_DELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: atom_1.TEST_DELEGATE_TX.hash,
                outputs: [
                    {
                        address: atom_1.TEST_DELEGATE_TX.validator,
                        amount: atom_1.TEST_DELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: atom_1.TEST_DELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: atom_1.TEST_DELEGATE_TX.gasBudget.amount[0].amount },
                type: 13,
            });
        });
        it('should explain a undelegate transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: atom_1.TEST_UNDELEGATE_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: atom_1.TEST_UNDELEGATE_TX.hash,
                outputs: [
                    {
                        address: atom_1.TEST_UNDELEGATE_TX.validator,
                        amount: atom_1.TEST_UNDELEGATE_TX.sendAmount,
                    },
                ],
                outputAmount: atom_1.TEST_UNDELEGATE_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: atom_1.TEST_UNDELEGATE_TX.gasBudget.amount[0].amount },
                type: 17,
            });
        });
        it('should explain a withdraw transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: atom_1.TEST_WITHDRAW_REWARDS_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: atom_1.TEST_WITHDRAW_REWARDS_TX.hash,
                outputs: [
                    {
                        address: atom_1.TEST_WITHDRAW_REWARDS_TX.validator,
                        amount: 'UNAVAILABLE',
                    },
                ],
                outputAmount: undefined,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: atom_1.TEST_WITHDRAW_REWARDS_TX.gasBudget.amount[0].amount },
                type: 15,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: atom_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: atom_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: atom_1.TEST_TX_WITH_MEMO.to,
                        amount: atom_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: atom_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: atom_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: atom_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await basecoin.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await basecoin.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        const transferInputsResponse = {
            address: atom_1.TEST_SEND_TX.recipient,
            amount: new bignumber_js_1.default(atom_1.TEST_SEND_TX.sendAmount).plus(atom_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
        };
        const transferOutputsResponse = {
            address: atom_1.TEST_SEND_TX.recipient,
            amount: atom_1.TEST_SEND_TX.sendAmount,
        };
        it('should parse a transfer transaction', async function () {
            const parsedTransaction = await basecoin.parseTransaction({ txHex: atom_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(src_1.Atom.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await basecoin
                .parseTransaction({ txHex: atom_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
    describe('Recover transactions: success path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = atom_1.wrwUser.destinationAddress;
        const coin = statics_1.coins.get('tatom');
        const testBalance = [
            {
                denom: 'uatom',
                amount: '150000',
            },
        ];
        const testAccountNumber = '725163';
        const testSequenceNumber = '8';
        const testBalanceDkls = [
            {
                denom: 'uatom',
                amount: '150000',
            },
        ];
        const testAccountNumberDkls = '755440';
        const testSequenceNumberDkls = '2';
        const testChainId = 'theta-testnet-001';
        beforeEach(() => {
            const accountBalance = sandBox.stub(src_1.Atom.prototype, 'getAccountBalance');
            accountBalance.withArgs(atom_1.wrwUser.senderAddress).resolves(testBalance);
            accountBalance.withArgs(atom_1.wrwUserDkls.senderAddress).resolves(testBalanceDkls);
            const accountDetails = sandBox.stub(src_1.Atom.prototype, 'getAccountDetails');
            accountDetails.withArgs(atom_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            accountDetails.withArgs(atom_1.wrwUserDkls.senderAddress).resolves([testAccountNumberDkls, testSequenceNumberDkls]);
            const chainId = sandBox.stub(src_1.Atom.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(function () {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should recover funds for non-bitgo recoveries', async function () {
            const res = await basecoin.recover({
                userKey: atom_1.wrwUser.userKey,
                backupKey: atom_1.wrwUser.backupKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                walletPassphrase: atom_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(basecoin.getAccountBalance);
            sandBox.assert.calledOnce(basecoin.getAccountDetails);
            sandBox.assert.calledOnce(basecoin.getChainId);
            const atomTxn = new abstract_cosmos_1.CosmosTransaction(coin, utils_1.default);
            atomTxn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const atomTxnJson = atomTxn.toJson();
            const sendMessage = atomTxnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const gasAmount = new bignumber_js_1.default(constants_1.GAS_AMOUNT);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
            should.equal(sendMessage.amount[0].denom, testBalance[0].denom);
        });
        it('should recover funds for non-bitgo recoveries - DKLS type', async function () {
            const res = await basecoin.recover({
                userKey: atom_1.wrwUserDkls.userKey,
                backupKey: atom_1.wrwUserDkls.backupKey,
                bitgoKey: atom_1.wrwUserDkls.bitgoKey,
                walletPassphrase: atom_1.wrwUserDkls.walletPassphrase,
                recoveryDestination: atom_1.wrwUserDkls.destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(basecoin.getAccountBalance);
            sandBox.assert.calledOnce(basecoin.getAccountDetails);
            sandBox.assert.calledOnce(basecoin.getChainId);
            const atomTxn = new abstract_cosmos_1.CosmosTransaction(coin, utils_1.default);
            atomTxn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const atomTxnJson = atomTxn.toJson();
            const sendMessage = atomTxnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalanceDkls[0].amount);
            const gasAmount = new bignumber_js_1.default(constants_1.GAS_AMOUNT);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
        });
        it('should recover funds for Unsigned Sweep Transaction', async function () {
            const res = await basecoin.recover({
                userKey: atom_1.wrwUser.userKey,
                backupKey: atom_1.wrwUser.backupKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                walletPassphrase: atom_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(basecoin.getAccountBalance);
            sandBox.assert.calledOnce(basecoin.getAccountDetails);
            sandBox.assert.calledOnce(basecoin.getChainId);
            const unsignedSweepTxnDeserialize = new abstract_cosmos_1.CosmosTransaction(coin, utils_1.default);
            unsignedSweepTxnDeserialize.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const unsignedSweepTxnJson = unsignedSweepTxnDeserialize.toJson();
            const sendMessage = unsignedSweepTxnJson.sendMessages[0].value;
            const balance = new bignumber_js_1.default(testBalance[0].amount);
            const gasAmount = new bignumber_js_1.default(constants_1.GAS_AMOUNT);
            const actualBalance = balance.minus(gasAmount);
            should.equal(sendMessage.amount[0].amount, actualBalance.toFixed());
        });
        it('should redelegate funds to new validator', async function () {
            const res = await basecoin.redelegate({
                userKey: atom_1.wrwUser.userKey,
                backupKey: atom_1.wrwUser.backupKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                walletPassphrase: atom_1.wrwUser.walletPassphrase,
                amountToRedelegate: '10000000000000000',
                validatorSrcAddress: 'cosmosvaloper1409te27da74uahh6hn0040x7l272hjs2padjuz',
                validatorDstAddress: 'cosmosvaloper183aycgtstp67r6s4vd7ts2npp2ckk4xah7rxj6',
            });
            res.should.not.be.empty();
            res.should.hasOwnProperty('serializedTx');
            sandBox.assert.calledOnce(basecoin.getChainId);
            const txn = new abstract_cosmos_1.CosmosTransaction(coin, utils_1.default);
            txn.enrichTransactionDetailsFromRawTransaction(res.serializedTx);
            const txnJson = txn.toJson();
            const redelegateMessage = txnJson.sendMessages[0].value;
            should.equal(redelegateMessage.validatorSrcAddress, 'cosmosvaloper1409te27da74uahh6hn0040x7l272hjs2padjuz');
            should.equal(redelegateMessage.validatorDstAddress, 'cosmosvaloper183aycgtstp67r6s4vd7ts2npp2ckk4xah7rxj6');
            should.equal(redelegateMessage.amount.amount, '10000000000000000');
        });
    });
    describe('Recover transactions: failure path', () => {
        const sandBox = sinon_1.default.createSandbox();
        const destinationAddress = atom_1.wrwUser.destinationAddress;
        const testZeroBalance = [
            {
                denom: 'uatom',
                amount: '0',
            },
        ];
        const testAccountNumber = '1234';
        const testSequenceNumber = '0';
        const testChainId = 'test-chain';
        beforeEach(() => {
            const accountBalance = sandBox.stub(src_1.Atom.prototype, 'getAccountBalance');
            accountBalance.withArgs(atom_1.wrwUser.senderAddress).resolves(testZeroBalance);
            const accountNumber = sandBox.stub(src_1.Atom.prototype, 'getAccountDetails');
            accountNumber.withArgs(atom_1.wrwUser.senderAddress).resolves([testAccountNumber, testSequenceNumber]);
            const chainId = sandBox.stub(src_1.Atom.prototype, 'getChainId');
            chainId.withArgs().resolves(testChainId);
            const deserializedEntChallenge = sdk_lib_mpc_1.EcdsaTypes.deserializeNtildeWithProofs(sdk_test_1.mockSerializedChallengeWithProofs);
            sinon_1.default.stub(sdk_lib_mpc_1.EcdsaRangeProof, 'generateNtilde').resolves(deserializedEntChallenge);
        });
        afterEach(function () {
            sandBox.restore();
            sinon_1.default.restore();
        });
        it('should throw error if backupkey is not present', async function () {
            await basecoin
                .recover({
                userKey: atom_1.wrwUser.userKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                walletPassphrase: atom_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing backupKey');
        });
        it('should throw error if userkey is not present', async function () {
            await basecoin
                .recover({
                backupKey: atom_1.wrwUser.backupKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                walletPassphrase: atom_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing userKey');
        });
        it('should throw error if wallet passphrase is not present', async function () {
            await basecoin
                .recover({
                userKey: atom_1.wrwUser.userKey,
                backupKey: atom_1.wrwUser.backupKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('missing wallet passphrase');
        });
        it('should throw error if there is no balance', async function () {
            await basecoin
                .recover({
                userKey: atom_1.wrwUser.userKey,
                backupKey: atom_1.wrwUser.backupKey,
                bitgoKey: atom_1.wrwUser.bitgoKey,
                walletPassphrase: atom_1.wrwUser.walletPassphrase,
                recoveryDestination: destinationAddress,
            })
                .should.rejectedWith('Did not have enough funds to recover');
        });
    });
});
