// @taquito/utils/dist/lib/taquito-utils.js
"use strict";
/**
 * @packageDocumentation
 * @module @taquito/utils
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.format = exports.POP_DST = exports.BLS12_381_DST = exports.verifySignature = exports.payloadLength = exports.PrefixV2 = exports.VERSION = void 0;
// Re-export everything from encoding
__exportStar(require("./encoding"), exports);
// Re-export everything from validators
__exportStar(require("./validators"), exports);
// Re-export version
var version_1 = require("./version");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_1.VERSION; } });
// Re-export constants
var constants_1 = require("./constants");
Object.defineProperty(exports, "PrefixV2", { enumerable: true, get: function () { return constants_1.PrefixV2; } });
Object.defineProperty(exports, "payloadLength", { enumerable: true, get: function () { return constants_1.payloadLength; } });
// Re-export from verify-signature
var verify_signature_1 = require("./verify-signature");
Object.defineProperty(exports, "verifySignature", { enumerable: true, get: function () { return verify_signature_1.verifySignature; } });
Object.defineProperty(exports, "BLS12_381_DST", { enumerable: true, get: function () { return verify_signature_1.BLS12_381_DST; } });
Object.defineProperty(exports, "POP_DST", { enumerable: true, get: function () { return verify_signature_1.POP_DST; } });
// Re-export errors
__exportStar(require("./errors"), exports);
// Re-export format
var format_1 = require("./format");
Object.defineProperty(exports, "format", { enumerable: true, get: function () { return format_1.format; } });
