// @bitgo/sdk-core/dist/src/bitgo/evm/evmUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateEvmKeyRingWalletParams = validateEvmKeyRingWalletParams;
exports.createEvmKeyRingWallet = createEvmKeyRingWallet;
const wallet_1 = require("../wallet");
const keychain_1 = require("../keychain");
/**
 * @param params - The wallet creation parameters
 * @param baseCoin - The base coin instance
 * @throws Error if validation fails
 * @returns boolean - true if validation passes
 */
function validateEvmKeyRingWalletParams(params, baseCoin) {
    if (!params.evmKeyRingReferenceWalletId)
        return false;
    if (typeof params.evmKeyRingReferenceWalletId !== 'string') {
        throw new Error('invalid evmKeyRingReferenceWalletId argument, expecting string');
    }
    if (!baseCoin.isEVM()) {
        throw new Error('evmKeyRingReferenceWalletId is only supported for EVM chains');
    }
    return true;
}
/**
 * Creates an EVM keyring wallet with shared keys from a reference wallet
 * @param params - The parameters for creating the EVM keyring wallet
 * @returns Promise<WalletWithKeychains> - The created wallet with its keychains
 */
async function createEvmKeyRingWallet(params) {
    const { label, evmKeyRingReferenceWalletId, bitgo, baseCoin } = params;
    // For EVM keyring wallets, this bypasses the normal key generation process since keys are shared via keyring
    const addWalletParams = {
        label,
        evmKeyRingReferenceWalletId,
    };
    const newWallet = await bitgo.post(baseCoin.url('/wallet/add')).send(addWalletParams).result();
    const userKeychain = baseCoin.keychains().get({ id: newWallet.keys[keychain_1.KeyIndices.USER] });
    const backupKeychain = baseCoin.keychains().get({ id: newWallet.keys[keychain_1.KeyIndices.BACKUP] });
    const bitgoKeychain = baseCoin.keychains().get({ id: newWallet.keys[keychain_1.KeyIndices.BITGO] });
    const [userKey, backupKey, bitgoKey] = await Promise.all([userKeychain, backupKeychain, bitgoKeychain]);
    const result = {
        wallet: new wallet_1.Wallet(bitgo, baseCoin, newWallet),
        userKeychain: userKey,
        backupKeychain: backupKey,
        bitgoKeychain: bitgoKey,
        responseType: 'WalletWithKeychains',
    };
    return result;
}
