// @taquito/signer/dist/taquito-signer.es6.js
import { openSecretBox } from '@stablelib/nacl';
import { b58DecodeAndCheckPrefix, PrefixV2, b58Encode, compareArrays, InvalidPublicKeyError, BLS12_381_DST, POP_DST, hex2buf, mergebuf, buf2hex } from '@taquito/utils';
import toBuffer from 'typedarray-to-buffer';
import { hash } from '@stablelib/blake2b';
import { generateKeyPairFromSeed, sign } from '@stablelib/ed25519';
import { secp256k1 } from '@noble/curves/secp256k1';
import { p256 } from '@noble/curves/nist.js';
import { ParameterValidationError, UnsupportedActionError, InvalidHexStringError, InvalidDerivationPathError, InvalidKeyError, ProhibitedActionError } from '@taquito/core';
import pbkdf2 from 'pbkdf2';
import * as Bip39 from 'bip39';
import { HMAC } from '@stablelib/hmac';
import { SHA512 } from '@stablelib/sha512';
import { p256 as p256$1 } from '@noble/curves/nist';
import BN from 'bn.js';
import { bls12_381 } from '@noble/curves/bls12-381';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol */


function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __classPrivateFieldGet(receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}

function __classPrivateFieldSet(receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var _EdKey_keyPair, _EdPublicKey_key;
/**
 * @description Provide signing logic for ed25519 curve based key (tz1)
 */
class EdKey {
    /**
     *
     * @param key Encoded private key
     * @param encrypted Is the private key encrypted
     * @param decrypt Decrypt function
     * @throws {@link InvalidKeyError}
     */
    constructor(key, decrypt) {
        _EdKey_keyPair.set(this, void 0);
        const tmp = b58DecodeAndCheckPrefix(key, [
            PrefixV2.Ed25519SecretKey,
            PrefixV2.Ed25519EncryptedSeed,
            PrefixV2.Ed25519Seed,
        ]);
        let [keyData] = tmp;
        const [, prefix] = tmp;
        if (prefix === PrefixV2.Ed25519SecretKey) {
            __classPrivateFieldSet(this, _EdKey_keyPair, {
                secretKey: keyData,
                publicKey: keyData.slice(32),
            }, "f");
        }
        else {
            if (prefix === PrefixV2.Ed25519EncryptedSeed) {
                if (decrypt !== undefined) {
                    keyData = decrypt(keyData);
                }
                else {
                    throw new Error('decryption function is not provided');
                }
            }
            __classPrivateFieldSet(this, _EdKey_keyPair, generateKeyPairFromSeed(keyData), "f");
        }
    }
    /**
     *
     * @param bytes Bytes to sign
     * @param bytesHash Blake2b hash of the bytes to sign
     */
    sign(bytes) {
        const hash$1 = hash(bytes, 32);
        const signature = sign(__classPrivateFieldGet(this, _EdKey_keyPair, "f").secretKey, hash$1);
        return {
            rawSignature: signature,
            sig: b58Encode(signature, PrefixV2.GenericSignature),
            prefixSig: b58Encode(signature, PrefixV2.Ed25519Signature),
        };
    }
    /**
     * @returns Encoded public key
     */
    publicKey() {
        return new EdPublicKey(__classPrivateFieldGet(this, _EdKey_keyPair, "f").publicKey);
    }
    /**
     * @returns Encoded private key
     */
    secretKey() {
        return b58Encode(__classPrivateFieldGet(this, _EdKey_keyPair, "f").secretKey, PrefixV2.Ed25519SecretKey);
    }
}
_EdKey_keyPair = new WeakMap();
class EdPublicKey {
    constructor(src) {
        _EdPublicKey_key.set(this, void 0);
        if (typeof src === 'string') {
            const [key, _] = b58DecodeAndCheckPrefix(src, [PrefixV2.Ed25519PublicKey]);
            __classPrivateFieldSet(this, _EdPublicKey_key, key, "f");
        }
        else {
            __classPrivateFieldSet(this, _EdPublicKey_key, src, "f");
        }
    }
    compare(other) {
        if (other instanceof EdPublicKey) {
            return compareArrays(this.bytes(), other.bytes());
        }
        else {
            throw new InvalidPublicKeyError('EdDSA key expected');
        }
    }
    hash() {
        return b58Encode(hash(__classPrivateFieldGet(this, _EdPublicKey_key, "f"), 20), PrefixV2.Ed25519PublicKeyHash);
    }
    bytes() {
        return __classPrivateFieldGet(this, _EdPublicKey_key, "f");
    }
    toProtocol() {
        const res = new Uint8Array(__classPrivateFieldGet(this, _EdPublicKey_key, "f").length + 1);
        res[0] = 0;
        res.set(__classPrivateFieldGet(this, _EdPublicKey_key, "f"), 1);
        return res;
    }
    toString() {
        return b58Encode(__classPrivateFieldGet(this, _EdPublicKey_key, "f"), PrefixV2.Ed25519PublicKey);
    }
}
_EdPublicKey_key = new WeakMap();

/**
 *  @category Error
 *  @description Error that indicates an invalid Mnemonic being passed or used
 */
class InvalidMnemonicError extends ParameterValidationError {
    constructor(mnemonic) {
        super();
        this.mnemonic = mnemonic;
        this.name = 'InvalidMnemonicError';
        this.message = `Invalid mnemonic "${mnemonic}"`;
    }
}
/**
 *  @category Error
 *  @description Error that indicates a curve with incorrect bit size being passed or used
 */
class InvalidBitSize extends ParameterValidationError {
    constructor(message) {
        super();
        this.message = message;
        this.name = 'InvalidBitSize';
    }
}
/**
 *  @category Error
 *  @description Error that indicates an unsupported cureve being passed or used
 */
class InvalidCurveError extends ParameterValidationError {
    constructor(message) {
        super();
        this.message = message;
        this.name = 'InvalidCurveError';
    }
}
/**
 *  @category Error
 *  @description Error that indicates a seed with invalid length being passed or used
 */
class InvalidSeedLengthError extends ParameterValidationError {
    constructor(seedLength) {
        super();
        this.seedLength = seedLength;
        this.name = 'InvalidSeedLengthError';
        this.message = `Invalid seed length "${seedLength}" expecting length between 16 to 64.`;
    }
}
/**
 *  @category Error
 *  @description Error that indicates a feature still under developement
 */
class ToBeImplemented extends UnsupportedActionError {
    constructor() {
        super();
        this.name = 'ToBeImplemented';
        this.message = 'This feature is under developement';
    }
}
/**
 *  @category Error
 *  @description Error that indicates an invalid passphrase being passed or used
 */
class InvalidPassphraseError extends ParameterValidationError {
    constructor(message) {
        super();
        this.message = message;
        this.name = 'InvalidPassphraseError';
    }
}

var _ECKey_keyPair, _ECPublicKey_key;
const pref = {
    p256: {
        pk: PrefixV2.P256PublicKey,
        sk: PrefixV2.P256SecretKey,
        pkh: PrefixV2.P256PublicKeyHash,
        sig: PrefixV2.P256Signature,
        tag: 2,
    },
    secp256k1: {
        pk: PrefixV2.Secp256k1PublicKey,
        sk: PrefixV2.Secp256k1SecretKey,
        pkh: PrefixV2.Secp256k1PublicKeyHash,
        sig: PrefixV2.Secp256k1Signature,
        tag: 1,
    },
};
/**
 * @description Provide signing logic for elliptic curve based key (tz2, tz3)
 */
class ECKey {
    /**
     *
     * @param key Encoded private key
     * @param decrypt Decrypt function
     * @throws {@link InvalidKeyError}
     */
    constructor(key, decrypt) {
        _ECKey_keyPair.set(this, void 0);
        const [keyData, prefix] = b58DecodeAndCheckPrefix(key, [
            PrefixV2.Secp256k1EncryptedSecretKey,
            PrefixV2.P256EncryptedSecretKey,
            PrefixV2.Secp256k1SecretKey,
            PrefixV2.P256SecretKey,
        ]);
        const [decKey, curve] = (() => {
            switch (prefix) {
                case PrefixV2.Secp256k1EncryptedSecretKey:
                case PrefixV2.P256EncryptedSecretKey:
                    if (decrypt === undefined) {
                        throw new Error('decryption function is not provided');
                    }
                    else {
                        return [
                            decrypt(keyData),
                            prefix === PrefixV2.Secp256k1EncryptedSecretKey ? 'secp256k1' : 'p256',
                        ];
                    }
                case PrefixV2.Secp256k1SecretKey:
                    return [keyData, 'secp256k1'];
                default:
                    return [keyData, 'p256'];
            }
        })();
        __classPrivateFieldSet(this, _ECKey_keyPair, {
            curve,
            secretKey: decKey,
            publicKey: curve === 'secp256k1'
                ? secp256k1.getPublicKey(decKey, true)
                : p256.getPublicKey(decKey, true),
        }, "f");
    }
    /**
     *
     * @param bytes Bytes to sign
     * @param bytesHash Blake2b hash of the bytes to sign
     */
    sign(bytes) {
        const hash$1 = hash(bytes, 32);
        let signature;
        if (__classPrivateFieldGet(this, _ECKey_keyPair, "f").curve === 'secp256k1') {
            signature = secp256k1
                .sign(hash$1, __classPrivateFieldGet(this, _ECKey_keyPair, "f").secretKey, {
                lowS: true, // Use canonical signatures (prevents malleability)
            })
                .toBytes('compact');
        }
        else {
            signature = p256
                .sign(hash$1, __classPrivateFieldGet(this, _ECKey_keyPair, "f").secretKey, {
                lowS: true, // Use canonical signatures (prevents malleability)
            })
                .toBytes('compact');
        }
        return {
            rawSignature: signature,
            sig: b58Encode(signature, PrefixV2.GenericSignature),
            prefixSig: b58Encode(signature, pref[__classPrivateFieldGet(this, _ECKey_keyPair, "f").curve].sig),
        };
    }
    /**
     * @returns Encoded public key
     */
    publicKey() {
        return new ECPublicKey(__classPrivateFieldGet(this, _ECKey_keyPair, "f").publicKey, __classPrivateFieldGet(this, _ECKey_keyPair, "f").curve);
    }
    /**
     * @returns Encoded private key
     */
    secretKey() {
        return b58Encode(__classPrivateFieldGet(this, _ECKey_keyPair, "f").secretKey, pref[__classPrivateFieldGet(this, _ECKey_keyPair, "f").curve].sk);
    }
}
_ECKey_keyPair = new WeakMap();
class ECPublicKey {
    constructor(src, curve) {
        _ECPublicKey_key.set(this, void 0);
        const [key, crv] = (() => {
            if (typeof src === 'string') {
                const [key, pre] = b58DecodeAndCheckPrefix(src, [
                    PrefixV2.Secp256k1PublicKey,
                    PrefixV2.P256PublicKey,
                ]);
                return [key, pre === PrefixV2.Secp256k1PublicKey ? 'secp256k1' : 'p256'];
            }
            else if (curve !== undefined) {
                return [src, curve];
            }
            else {
                throw new InvalidCurveError('missing curve type');
            }
        })();
        // For ECPublicKey, we don't need the private key, so we pass an empty array
        // The public key is stored separately
        __classPrivateFieldSet(this, _ECPublicKey_key, key, "f");
        this.curve = crv;
    }
    compare(other) {
        if (other instanceof ECPublicKey) {
            if (this.curve === other.curve) {
                const compress = this.curve === 'secp256k1';
                return compareArrays(this.bytes(compress), other.bytes(compress));
            }
            else if (this.curve === 'secp256k1') {
                return -1;
            }
            else {
                return 1;
            }
        }
        else {
            throw new InvalidPublicKeyError('ECDSA key expected');
        }
    }
    hash() {
        const key = this.bytes();
        return b58Encode(hash(key, 20), pref[this.curve].pkh);
    }
    bytes(compress = true) {
        // @noble/curves supports both compressed and uncompressed formats
        // We need to convert the stored public key to the requested format
        if (this.curve === 'secp256k1') {
            // For secp256k1, we need to get the public key from the stored bytes and convert format
            const point = secp256k1.Point.fromHex(__classPrivateFieldGet(this, _ECPublicKey_key, "f"));
            return point.toBytes(compress);
        }
        else {
            // For p256, we need to get the public key from the stored bytes and convert format
            const point = p256.Point.fromHex(__classPrivateFieldGet(this, _ECPublicKey_key, "f"));
            return point.toBytes(compress);
        }
    }
    toProtocol() {
        const key = this.bytes();
        const res = new Uint8Array(key.length + 1);
        res[0] = pref[this.curve].tag;
        res.set(key, 1);
        return res;
    }
    toString() {
        const key = this.bytes();
        return b58Encode(key, pref[this.curve].pk);
    }
}
_ECPublicKey_key = new WeakMap();

const Hard = 0x80000000;

function parseHex(s) {
    const res = [];
    for (let i = 0; i < s.length; i += 2) {
        const ss = s.slice(i, i + 2);
        const x = parseInt(ss, 16);
        if (Number.isNaN(x)) {
            throw new InvalidHexStringError(ss);
        }
        res.push(x);
    }
    return new Uint8Array(res);
}

/* eslint-disable @typescript-eslint/no-this-alias */
// MinSeedSize is the minimal allowed seed byte length
const minSeedSize$1 = 16;
// MaxSeedSize is the maximal allowed seed byte length
const maxSeedSize$1 = 64;
const ed25519Key = 'ed25519 seed';
let PrivateKey$1 = class PrivateKey {
    /**
     *
     * @param priv generated keypair 0->32 private key 32->n public key
     * @param chainCode new HMAC hash with new key
     */
    constructor(priv, chainCode) {
        this.priv = priv;
        this.chainCode = chainCode;
    }
    /**
     *
     * @param seedSrc result of Bip39.mnemonicToSeed
     * @returns instance of PrivateKey
     * @throws {@link InvalidSeedLengthError}
     */
    static fromSeed(seedSrc) {
        const seed = typeof seedSrc === 'string' ? parseHex(seedSrc) : seedSrc;
        if (seed.length < minSeedSize$1 || seed.length > maxSeedSize$1) {
            throw new InvalidSeedLengthError(seed.length);
        }
        const key = new TextEncoder().encode(ed25519Key);
        const sum = new HMAC(SHA512, key).update(seed).digest();
        return new PrivateKey(generateKeyPairFromSeed(sum.subarray(0, 32)).secretKey, sum.subarray(32));
    }
    /**
     *
     * @returns slice(0, 32) of current priv for new seed for next derived priv
     */
    seed() {
        return this.priv.subarray(0, 32);
    }
    /**
     * @index current derivation path item ie: 1729'
     * @returns derivation path child of original private key pair
     */
    derive(index) {
        if ((index & Hard) === 0) {
            throw new InvalidDerivationPathError(index.toString(), ': Non-hardened derivation path.');
        }
        const data = new Uint8Array(37);
        data.set(this.seed(), 1);
        new DataView(data.buffer).setUint32(33, index);
        const sum = new HMAC(SHA512, this.chainCode).update(data).digest();
        return new PrivateKey(generateKeyPairFromSeed(sum.subarray(0, 32)).secretKey, sum.subarray(32));
    }
    /**
     * @param path array of numbers pre adjusted for hardened paths ie: 44' -> 2^31 + 44
     * @returns final child of full derivation path private key pair
     */
    derivePath(path) {
        let key = this;
        for (const index of path) {
            key = key.derive(index);
        }
        return key;
    }
};

var ed25519 = /*#__PURE__*/Object.freeze({
    __proto__: null,
    PrivateKey: PrivateKey$1
});

/* eslint-disable @typescript-eslint/no-this-alias */
const seedKey = {
    p256: 'Nist256p1 seed',
    secp256k1: 'Bitcoin seed',
};
// MinSeedSize is the minimal allowed seed byte length
const minSeedSize = 16;
// MaxSeedSize is the maximal allowed seed byte length
const maxSeedSize = 64;
class PrivateKey {
    /**
     *
     * @param priv { secretKey: BN, curve: 'p256' | 'secp256k1' }
     * @param chainCode slice 32->n HMAC hash key and seedkey (first instance curve default seedKey. after hmac value slice 32->n)
     */
    constructor(priv, chainCode) {
        this.priv = priv;
        this.chainCode = chainCode;
        const privateKeyBytes = priv.secretKey.toArray('be', 32);
        const privateKeyUint8 = new Uint8Array(privateKeyBytes);
        const publicKey = priv.curve === 'secp256k1'
            ? secp256k1.getPublicKey(privateKeyUint8, true) // compressed public key
            : p256$1.getPublicKey(privateKeyUint8, true); // compressed public key
        this.keyPair = {
            curve: priv.curve,
            secretKey: priv.secretKey,
            publicKey: publicKey,
        };
    }
    /**
     * @param seedSrc result of Bip39.mnemonicToSeed
     * @param curve known supported curve p256 or secp256k1
     * @returns instance of PrivateKey non-HD keys derived
     * @throws {@link InvalidBitSize} | {@link InvalidCurveError} | {@link InvalidSeedLengthError}
     */
    static fromSeed(seedSrc, curve) {
        let seed = typeof seedSrc === 'string' ? parseHex(seedSrc) : seedSrc;
        if (seed.length < minSeedSize || seed.length > maxSeedSize) {
            throw new InvalidSeedLengthError(seed.length);
        }
        if (!Object.prototype.hasOwnProperty.call(seedKey, curve)) {
            throw new InvalidCurveError(`Unsupported curve "${curve}" expecting either "p256" or "secp256k1"`);
        }
        // Get curve order for validation
        let curveOrder;
        if (curve === 'secp256k1') {
            curveOrder = new BN(secp256k1.Point.CURVE().n.toString());
        }
        else {
            curveOrder = new BN(p256$1.Point.CURVE().n.toString());
        }
        if (curveOrder.bitLength() !== 256) {
            throw new InvalidBitSize(`Invalid curve "${curve}" with bit size "${curveOrder.bitLength()}" expecting bit size "256"`);
        }
        const key = new TextEncoder().encode(seedKey[curve]);
        let d = null;
        let chain = new Uint8Array();
        let i = 0;
        while (i === 0) {
            const sum = new HMAC(SHA512, key).update(seed).digest();
            d = new BN(sum.subarray(0, 32));
            chain = sum.subarray(32);
            if (d.isZero() || d.cmp(curveOrder) >= 0) {
                seed = sum;
            }
            else {
                i++;
            }
        }
        return new PrivateKey({ curve, secretKey: d }, chain);
    }
    /**
     *
     * @param index derivation path item pre-hardened if applicable ie: 44' -> 2^31 + 44
     * @returns child PrivateKey of the current PrivateKey
     */
    derive(index) {
        const data = new Uint8Array(37);
        if ((index & Hard) !== 0) {
            // hardened derivation
            data.set(this.keyPair.secretKey.toArray(), 1);
        }
        else {
            data.set(this.keyPair.publicKey, 0);
        }
        new DataView(data.buffer).setUint32(33, index);
        let d = new BN(0);
        let chain = new Uint8Array();
        let i = 0;
        while (i === 0) {
            const sum = new HMAC(SHA512, this.chainCode).update(data).digest();
            d = new BN(sum.subarray(0, 32));
            chain = sum.subarray(32);
            // Get curve order for comparison
            let curveOrder;
            if (this.keyPair.curve === 'secp256k1') {
                curveOrder = new BN(secp256k1.Point.CURVE().n.toString());
            }
            else {
                curveOrder = new BN(p256$1.Point.CURVE().n.toString());
            }
            if (d.cmp(curveOrder) < 0) {
                d = d.add(this.keyPair.secretKey).mod(curveOrder);
                if (!d.isZero()) {
                    i++;
                }
            }
            data.set(chain, 1);
            data[0] = 1;
        }
        return new PrivateKey({ curve: this.keyPair.curve, secretKey: d }, chain);
    }
    /**
     *
     * @param path pre-hardened (if applicable) derivation path items ie 44'/1729'/0/0 -> 2^31 + 44/2^31 + 1729/0/0
     * @returns final child of the full HD keys derivation
     */
    derivePath(path) {
        let key = this;
        for (const x of path) {
            key = key.derive(x);
        }
        return key;
    }
    /**
     *
     * @returns Uint8Array (if contains a private key)
     * @throws {@link InvalidKeyError}
     */
    bytes() {
        if (!this.keyPair.secretKey) {
            throw new InvalidKeyError('missing private key');
        }
        // pad to 32 bytes as toArray() length argument seems to be ignored (BN bug)
        const src = this.keyPair.secretKey.toArray();
        const out = new Uint8Array(32);
        out.set(src, out.length - src.length);
        return out;
    }
}

var ecdsa = /*#__PURE__*/Object.freeze({
    __proto__: null,
    PrivateKey: PrivateKey
});

class Path extends Array {
    static from(iterable) {
        return super.from(iterable).map((x) => x >>> 0);
    }
    /**
     *
     * @param s derivation path eg: 44'/1729'/0'/0'
     * @returns applied hardened values
     */
    static fromString(s) {
        if (s.length === 0) {
            return new Path();
        }
        let parts = s.split('/');
        const out = [];
        if (parts[0] === 'm') {
            parts = parts.slice(1);
        }
        for (let p of parts) {
            if (p.length === 0) {
                throw new InvalidDerivationPathError(s, `: Invalid BIP32 path`);
            }
            let h = 0;
            const last = p[p.length - 1];
            if (last === "'" || last === 'h' || last === 'H') {
                h = Hard;
                p = p.slice(0, p.length - 1);
            }
            const index = (parseInt(p, 10) | h) >>> 0;
            out.push(index);
        }
        return Path.from(out);
    }
}

// bip32 when supported add to @param curve bip25519
/**
 *
 * @param seed bip39.mnemonicToSeed
 * @param derivationPath Tezos Requirement 44'/1729' for HD key address default 44'/1729'/0'/0'
 * @param curve 'ed25519' | 'secp256k1' | 'p256''
 * @returns final Derivation of HD keys tezos Secret key
 * @throws {@link InvalidCurveError} | {@link ToBeImplemented}
 */
const generateSecretKey = (seed, derivationPath, curve) => {
    const path = Path.fromString(derivationPath);
    let node;
    switch (curve) {
        case 'ed25519': {
            node = PrivateKey$1.fromSeed(seed).derivePath(path);
            const sk = b58Encode(node.seed().slice(0, 32), PrefixV2.Ed25519Seed);
            return sk;
        }
        case 'secp256k1':
        case 'p256': {
            const prefixType = curve === 'secp256k1' ? PrefixV2.Secp256k1SecretKey : PrefixV2.P256SecretKey;
            let privKey = PrivateKey.fromSeed(seed, curve);
            privKey = privKey.derivePath(path);
            const uint8arr = new Uint8Array(privKey.keyPair.secretKey.toArray());
            const sk = b58Encode(uint8arr, prefixType);
            return sk;
        }
        case 'bip25519': {
            throw new ToBeImplemented();
        }
        default: {
            throw new InvalidCurveError(`Unsupported curve "${curve}" expecting one of the following "ed25519", "secp256k1", "p256"`);
        }
    }
};

function isPOP(k) {
    return 'provePossession' in k;
}

var _BLSKey_key, _BLSKey_publicKey, _BLSPublicKey_key;
const bls = bls12_381.longSignatures; // AKA MinPK
class BLSKey {
    constructor(key, decrypt) {
        _BLSKey_key.set(this, void 0);
        _BLSKey_publicKey.set(this, void 0);
        const tmp = b58DecodeAndCheckPrefix(key, [
            PrefixV2.BLS12_381EncryptedSecretKey,
            PrefixV2.BLS12_381SecretKey,
        ]);
        let [keyData] = tmp;
        const [, prefix] = tmp;
        if (prefix === PrefixV2.BLS12_381EncryptedSecretKey) {
            if (decrypt !== undefined) {
                keyData = decrypt(keyData);
            }
            else {
                throw new Error('decryption function is not provided');
            }
        }
        __classPrivateFieldSet(this, _BLSKey_key, keyData, "f");
        __classPrivateFieldSet(this, _BLSKey_publicKey, bls.getPublicKey(this.sk()).toBytes(), "f");
    }
    sk() {
        return new Uint8Array(__classPrivateFieldGet(this, _BLSKey_key, "f")).reverse();
    }
    signDst(message, dst) {
        const point = bls.hash(message, dst);
        const sig = bls.sign(point, this.sk()).toBytes();
        return {
            rawSignature: sig,
            sig: b58Encode(sig, PrefixV2.GenericSignature),
            prefixSig: b58Encode(sig, PrefixV2.BLS12_381Signature),
        };
    }
    sign(message) {
        return this.signDst(message, BLS12_381_DST);
    }
    provePossession() {
        return this.signDst(__classPrivateFieldGet(this, _BLSKey_publicKey, "f"), POP_DST);
    }
    publicKey() {
        return new BLSPublicKey(__classPrivateFieldGet(this, _BLSKey_publicKey, "f"));
    }
    secretKey() {
        return b58Encode(__classPrivateFieldGet(this, _BLSKey_key, "f"), PrefixV2.BLS12_381SecretKey);
    }
}
_BLSKey_key = new WeakMap(), _BLSKey_publicKey = new WeakMap();
class BLSPublicKey {
    constructor(src) {
        _BLSPublicKey_key.set(this, void 0);
        if (typeof src === 'string') {
            const [key, _] = b58DecodeAndCheckPrefix(src, [PrefixV2.BLS12_381PublicKey]);
            __classPrivateFieldSet(this, _BLSPublicKey_key, key, "f");
        }
        else {
            __classPrivateFieldSet(this, _BLSPublicKey_key, src, "f");
        }
    }
    compare(other) {
        if (other instanceof BLSPublicKey) {
            return compareArrays(this.bytes(), other.bytes());
        }
        else {
            throw new InvalidPublicKeyError('BLS key expected');
        }
    }
    hash() {
        return b58Encode(hash(__classPrivateFieldGet(this, _BLSPublicKey_key, "f"), 20), PrefixV2.BLS12_381PublicKeyHash);
    }
    bytes() {
        return __classPrivateFieldGet(this, _BLSPublicKey_key, "f");
    }
    toProtocol() {
        const res = new Uint8Array(__classPrivateFieldGet(this, _BLSPublicKey_key, "f").length + 1);
        res[0] = 3;
        res.set(__classPrivateFieldGet(this, _BLSPublicKey_key, "f"), 1);
        return res;
    }
    toString() {
        return b58Encode(__classPrivateFieldGet(this, _BLSPublicKey_key, "f"), PrefixV2.BLS12_381PublicKey);
    }
}
_BLSPublicKey_key = new WeakMap();

var _InMemorySigner_key;
/**
 * @description A local implementation of the signer. Will represent a Tezos account and be able to produce signature in its behalf
 *
 * @warn If running in production and dealing with tokens that have real value, it is strongly recommended to use a HSM backed signer so that private key material is not stored in memory or on disk
 * @throws {@link InvalidMnemonicError}
 */
class InMemorySigner {
    static fromFundraiser(email, password, mnemonic) {
        if (!Bip39.validateMnemonic(mnemonic)) {
            throw new InvalidMnemonicError(mnemonic);
        }
        const seed = Bip39.mnemonicToSeedSync(mnemonic, `${email}${password}`);
        const key = b58Encode(seed.subarray(0, 32), PrefixV2.Ed25519Seed);
        return new InMemorySigner(key);
    }
    static fromSecretKey(key, passphrase) {
        return __awaiter(this, void 0, void 0, function* () {
            return new InMemorySigner(key, passphrase);
        });
    }
    /**
     *
     * @description Instantiation of an InMemorySigner instance from a mnemonic
     * @param mnemonic 12-24 word mnemonic
     * @param password password used to encrypt the mnemonic to seed value
     * @param derivationPath default 44'/1729'/0'/0' (44'/1729' mandatory)
     * @param curve currently only supported for tz1, tz2, tz3 addresses. soon bip25519
     * @returns InMemorySigner
     * @throws {@link InvalidMnemonicError}
     */
    static fromMnemonic({ mnemonic, password = '', derivationPath = "44'/1729'/0'/0'", curve = 'ed25519', }) {
        // check if curve is defined if not default tz1
        if (!Bip39.validateMnemonic(mnemonic)) {
            // avoiding exposing mnemonic again in case of mistake making invalid
            throw new InvalidMnemonicError(mnemonic);
        }
        const seed = Bip39.mnemonicToSeedSync(mnemonic, password);
        const sk = generateSecretKey(seed, derivationPath, curve);
        return new InMemorySigner(sk);
    }
    /**
     *
     * @param key Encoded private key
     * @param passphrase Passphrase to decrypt the private key if it is encrypted
     * @throws {@link InvalidKeyError}
     *
     */
    constructor(key, passphrase) {
        _InMemorySigner_key.set(this, void 0);
        const keyPrefixes = [
            PrefixV2.Ed25519EncryptedSeed,
            PrefixV2.Ed25519Seed,
            PrefixV2.Ed25519SecretKey,
            PrefixV2.Secp256k1EncryptedSecretKey,
            PrefixV2.Secp256k1SecretKey,
            PrefixV2.P256EncryptedSecretKey,
            PrefixV2.P256SecretKey,
            PrefixV2.BLS12_381EncryptedSecretKey,
            PrefixV2.BLS12_381SecretKey,
        ];
        const pre = (() => {
            try {
                const [, pre] = b58DecodeAndCheckPrefix(key, keyPrefixes);
                return pre;
            }
            catch (_a) {
                throw new InvalidKeyError(`Invalid private key, expecting one of the following prefixes '${keyPrefixes}'.`);
            }
        })();
        const encrypted = pre === PrefixV2.Ed25519EncryptedSeed ||
            pre === PrefixV2.Secp256k1EncryptedSecretKey ||
            pre === PrefixV2.P256EncryptedSecretKey ||
            pre === PrefixV2.BLS12_381EncryptedSecretKey;
        let decrypt;
        if (encrypted) {
            if (!passphrase) {
                throw new InvalidPassphraseError('No passphrase provided to decrypt encrypted key');
            }
            decrypt = (data) => {
                const salt = toBuffer(data.slice(0, 8));
                const encryptedSk = data.slice(8);
                const encryptionKey = pbkdf2.pbkdf2Sync(passphrase, salt, 32768, 32, 'sha512');
                const res = openSecretBox(new Uint8Array(encryptionKey), new Uint8Array(24), new Uint8Array(encryptedSk));
                if (!res) {
                    throw new Error("can't decrypt secret key");
                }
                return res;
            };
        }
        switch (pre) {
            case PrefixV2.Ed25519EncryptedSeed:
            case PrefixV2.Ed25519Seed:
            case PrefixV2.Ed25519SecretKey:
                __classPrivateFieldSet(this, _InMemorySigner_key, new EdKey(key, decrypt), "f");
                break;
            case PrefixV2.Secp256k1EncryptedSecretKey:
            case PrefixV2.Secp256k1SecretKey:
            case PrefixV2.P256EncryptedSecretKey:
            case PrefixV2.P256SecretKey:
                __classPrivateFieldSet(this, _InMemorySigner_key, new ECKey(key, decrypt), "f");
                break;
            case PrefixV2.BLS12_381EncryptedSecretKey:
            case PrefixV2.BLS12_381SecretKey:
                __classPrivateFieldSet(this, _InMemorySigner_key, new BLSKey(key, decrypt), "f");
                break;
        }
    }
    /**
     *
     * @param bytes Bytes to sign
     * @param watermark Watermark to append to the bytes
     */
    sign(message, watermark) {
        return __awaiter(this, void 0, void 0, function* () {
            const msg = typeof message == 'string' ? hex2buf(message) : message;
            const watermarkMsg = watermark !== undefined ? mergebuf(watermark, msg) : msg;
            const { rawSignature, sig: signature, prefixSig: prefixedSignature, } = yield __classPrivateFieldGet(this, _InMemorySigner_key, "f").sign(watermarkMsg);
            return {
                bytes: buf2hex(msg),
                sig: signature,
                prefixSig: prefixedSignature,
                sbytes: buf2hex(mergebuf(msg, 
                // bls only Signature_prefix ff03 ref:https://octez.tezos.com/docs/shell/p2p_api.html#signature-prefix-tag-255 & https://octez.tezos.com/docs/shell/p2p_api.html#bls-prefix-tag-3
                isPOP(__classPrivateFieldGet(this, _InMemorySigner_key, "f")) ? mergebuf(new Uint8Array([255, 3]), rawSignature) : rawSignature)),
            };
        });
    }
    provePossession() {
        return __awaiter(this, void 0, void 0, function* () {
            if (isPOP(__classPrivateFieldGet(this, _InMemorySigner_key, "f"))) {
                return __classPrivateFieldGet(this, _InMemorySigner_key, "f").provePossession();
            }
            else {
                throw new ProhibitedActionError('Only BLS keys can prove possession');
            }
        });
    }
    get canProvePossession() {
        return isPOP(__classPrivateFieldGet(this, _InMemorySigner_key, "f"));
    }
    /**
     * @returns Encoded public key
     */
    publicKey() {
        return Promise.resolve(String(__classPrivateFieldGet(this, _InMemorySigner_key, "f").publicKey()));
    }
    /**
     * @returns Encoded public key hash
     */
    publicKeyHash() {
        return Promise.resolve(__classPrivateFieldGet(this, _InMemorySigner_key, "f").publicKey().hash());
    }
    /**
     * @returns Encoded private key
     */
    secretKey() {
        return Promise.resolve(__classPrivateFieldGet(this, _InMemorySigner_key, "f").secretKey());
    }
}
_InMemorySigner_key = new WeakMap();
function publicKeyFromString(src) {
    const [keyData, pre] = b58DecodeAndCheckPrefix(src, [
        PrefixV2.Ed25519PublicKey,
        PrefixV2.Secp256k1PublicKey,
        PrefixV2.P256PublicKey,
        PrefixV2.BLS12_381PublicKey,
    ]);
    switch (pre) {
        case PrefixV2.Ed25519PublicKey:
            return new EdPublicKey(keyData);
        case PrefixV2.Secp256k1PublicKey:
            return new ECPublicKey(keyData, 'secp256k1');
        case PrefixV2.P256PublicKey:
            return new ECPublicKey(keyData, 'p256');
        case PrefixV2.BLS12_381PublicKey:
            return new BLSPublicKey(keyData);
    }
}

// IMPORTANT: THIS FILE IS AUTO GENERATED! DO NOT MANUALLY EDIT OR CHECKIN!
const VERSION = {
    "commitHash": "c77fe4b0989665d8b5cfd15a7cc977499021f6fd",
    "version": "23.1.0"
};

export { ecdsa as ECDSA, ed25519 as Ed25519, Hard, InMemorySigner, InvalidPassphraseError, Path, VERSION, generateSecretKey, publicKeyFromString };
