// @bitgo/sdk-coin-iota/dist/test/unit/helpers/testHelpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.testData = void 0;
exports.createBasicTransferBuilder = createBasicTransferBuilder;
exports.createTransferBuilderWithGas = createTransferBuilderWithGas;
exports.createTransferBuilderWithSponsor = createTransferBuilderWithSponsor;
exports.createBuilderWithRecipients = createBuilderWithRecipients;
exports.createBuilderWithGasObjectsOnly = createBuilderWithGasObjectsOnly;
exports.createBuilderWithGasObjects = createBuilderWithGasObjects;
exports.assertBasicTransactionProperties = assertBasicTransactionProperties;
exports.assertGasData = assertGasData;
exports.assertInputsAndOutputs = assertInputsAndOutputs;
exports.assertValidRawTransaction = assertValidRawTransaction;
exports.createRecipients = createRecipients;
exports.getFactory = getFactory;
const statics_1 = require("@bitgo/statics");
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/iota"));
exports.testData = testData;
const should_1 = __importDefault(require("should"));
/**
 * Common test helpers and fixtures for IOTA transaction tests.
 * Reduces duplication and improves test readability.
 */
const factory = new src_1.TransactionBuilderFactory(statics_1.coins.get('tiota'));
/**
 * Creates a basic transfer builder with sender, recipients, and payment objects.
 */
function createBasicTransferBuilder() {
    return factory
        .getTransferBuilder()
        .sender(testData.sender.address)
        .recipients(testData.recipients)
        .paymentObjects(testData.paymentObjects);
}
/**
 * Creates a transfer builder with gas data for real (non-simulate) transactions.
 */
function createTransferBuilderWithGas() {
    return createBasicTransferBuilder().gasData(testData.gasData);
}
/**
 * Creates a transfer builder with a gas sponsor.
 */
function createTransferBuilderWithSponsor() {
    return createTransferBuilderWithGas().gasSponsor(testData.gasSponsor.address);
}
/**
 * Creates a builder with custom recipients.
 */
function createBuilderWithRecipients(recipients) {
    return factory
        .getTransferBuilder()
        .sender(testData.sender.address)
        .recipients(recipients)
        .paymentObjects(testData.paymentObjects)
        .gasData(testData.gasData);
}
/**
 * Creates a builder that uses gas objects for payment (no payment objects).
 */
function createBuilderWithGasObjectsOnly() {
    return factory
        .getTransferBuilder()
        .sender(testData.sender.address)
        .recipients(testData.recipients)
        .gasData(testData.gasData);
}
/**
 * Creates a builder with custom gas objects.
 */
function createBuilderWithGasObjects(gasObjects) {
    return factory.getTransferBuilder().sender(testData.sender.address).recipients(testData.recipients).gasData({
        gasBudget: testData.GAS_BUDGET,
        gasPrice: testData.GAS_PRICE,
        gasPaymentObjects: gasObjects,
    });
}
/**
 * Asserts that a transaction has the expected basic properties.
 */
function assertBasicTransactionProperties(tx) {
    should_1.default.exist(tx);
    should_1.default.equal(tx.sender, testData.sender.address);
    should_1.default.deepEqual(tx.recipients, testData.recipients);
    should_1.default.deepEqual(tx.paymentObjects, testData.paymentObjects);
}
/**
 * Asserts that a transaction has valid gas data.
 */
function assertGasData(tx) {
    should_1.default.equal(tx.gasBudget, testData.GAS_BUDGET);
    should_1.default.equal(tx.gasPrice, testData.GAS_PRICE);
    should_1.default.deepEqual(tx.gasPaymentObjects, testData.gasPaymentObjects);
}
/**
 * Asserts that a transaction has the expected inputs and outputs.
 */
function assertInputsAndOutputs(tx, expectedRecipients) {
    const totalAmount = expectedRecipients.reduce((sum, r) => sum + Number(r.amount), 0);
    tx.inputs.length.should.equal(1);
    tx.inputs[0].should.deepEqual({
        address: testData.sender.address,
        value: totalAmount.toString(),
        coin: 'tiota',
    });
    tx.outputs.length.should.equal(expectedRecipients.length);
    expectedRecipients.forEach((recipient, index) => {
        tx.outputs[index].should.deepEqual({
            address: recipient.address,
            value: recipient.amount,
            coin: 'tiota',
        });
    });
}
/**
 * Verifies that a raw transaction is valid.
 */
async function assertValidRawTransaction(tx) {
    const rawTx = await tx.toBroadcastFormat();
    should_1.default.exist(rawTx);
    should_1.default.equal(typeof rawTx, 'string');
    should_1.default.equal(rawTx.length > 0, true);
}
/**
 * Creates an array of recipients with the specified count and amount.
 */
function createRecipients(count, amount = '1000') {
    return Array.from({ length: count }, (_, i) => ({
        address: testData.addresses.validAddresses[i % testData.addresses.validAddresses.length],
        amount,
    }));
}
/**
 * Gets the transaction factory instance.
 */
function getFactory() {
    return factory;
}
