// @bitgo/sdk-coin-zeta/dist/test/unit/transaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const encoding_1 = require("@cosmjs/encoding");
const should_1 = __importDefault(require("should"));
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const utils_1 = __importDefault(require("../../src/lib/utils"));
const testData = __importStar(require("../resources/zeta"));
describe('Zeta Transaction', () => {
    let tx;
    const config = statics_1.coins.get('tzeta');
    beforeEach(() => {
        tx = new abstract_cosmos_1.CosmosTransaction(config, utils_1.default);
    });
    describe('Empty transaction', () => {
        it('should throw empty transaction', function () {
            should_1.default.throws(() => tx.toBroadcastFormat(), 'Empty transaction');
        });
    });
    describe('From raw transaction', () => {
        it('should build a transfer from raw signed base64', function () {
            tx.enrichTransactionDetailsFromRawTransaction(testData.TEST_SEND_TX.signedTxBase64);
            const json = tx.toJson();
            should_1.default.equal(json.sequence, testData.TEST_SEND_TX.sequence);
            should_1.default.deepEqual(json.gasBudget, testData.TEST_SEND_TX.gasBudget);
            should_1.default.equal(json.publicKey, (0, sdk_core_1.toHex)((0, encoding_1.fromBase64)(testData.TEST_SEND_TX.pubKey)));
            should_1.default.equal(json.sendMessages[0].value.toAddress, testData.TEST_SEND_TX.sendMessage.value.toAddress);
            should_1.default.deepEqual(json.sendMessages[0].value.amount, testData.TEST_SEND_TX.sendMessage.value.amount);
            should_1.default.equal(Buffer.from(json.signature).toString('base64'), testData.TEST_SEND_TX.signature);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.loadInputsAndOutputs();
            should_1.default.deepEqual(tx.inputs, [
                {
                    address: testData.TEST_SEND_TX.sender,
                    value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                    coin: 'tzeta',
                },
            ]);
            should_1.default.deepEqual(tx.outputs, [
                {
                    address: testData.TEST_SEND_TX.sendMessage.value.toAddress,
                    value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                    coin: 'tzeta',
                },
            ]);
        });
        it('should build a transfer from raw signed hex', function () {
            tx.enrichTransactionDetailsFromRawTransaction((0, sdk_core_1.toHex)((0, encoding_1.fromBase64)(testData.TEST_SEND_TX.signedTxBase64)));
            const json = tx.toJson();
            should_1.default.equal(json.sequence, testData.TEST_SEND_TX.sequence);
            should_1.default.deepEqual(json.gasBudget, testData.TEST_SEND_TX.gasBudget);
            should_1.default.equal(json.publicKey, (0, sdk_core_1.toHex)((0, encoding_1.fromBase64)(testData.TEST_SEND_TX.pubKey)));
            should_1.default.equal(json.sendMessages[0].value.toAddress, testData.TEST_SEND_TX.sendMessage.value.toAddress);
            should_1.default.deepEqual(json.sendMessages[0].value.amount, testData.TEST_SEND_TX.sendMessage.value.amount);
            should_1.default.equal(Buffer.from(json.signature).toString('base64'), testData.TEST_SEND_TX.signature);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.Send);
            tx.loadInputsAndOutputs();
            should_1.default.deepEqual(tx.inputs, [
                {
                    address: testData.TEST_SEND_TX.sender,
                    value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                    coin: 'tzeta',
                },
            ]);
            should_1.default.deepEqual(tx.outputs, [
                {
                    address: testData.TEST_SEND_TX.sendMessage.value.toAddress,
                    value: testData.TEST_SEND_TX.sendMessage.value.amount[0].amount,
                    coin: 'tzeta',
                },
            ]);
        });
        it('should build a delegate txn from raw signed base64', function () {
            tx.enrichTransactionDetailsFromRawTransaction(testData.TEST_DELEGATE_TX.signedTxBase64);
            const json = tx.toJson();
            should_1.default.equal(json.sequence, testData.TEST_DELEGATE_TX.sequence);
            should_1.default.deepEqual(json.gasBudget, testData.TEST_DELEGATE_TX.gasBudget);
            should_1.default.equal(Buffer.from(json.publicKey, 'hex').toString('base64'), testData.TEST_DELEGATE_TX.pubKey);
            should_1.default.equal(json.sendMessages[0].value.validatorAddress, testData.TEST_DELEGATE_TX.sendMessage.value.validatorAddress);
            should_1.default.deepEqual(json.sendMessages[0].value.amount, testData.TEST_DELEGATE_TX.sendMessage.value.amount);
            should_1.default.equal(Buffer.from(json.signature).toString('base64'), testData.TEST_DELEGATE_TX.signature);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingActivate);
            tx.loadInputsAndOutputs();
            should_1.default.deepEqual(tx.inputs, [
                {
                    address: testData.TEST_DELEGATE_TX.delegator,
                    value: testData.TEST_DELEGATE_TX.sendMessage.value.amount.amount,
                    coin: 'tzeta',
                },
            ]);
            should_1.default.deepEqual(tx.outputs, [
                {
                    address: testData.TEST_DELEGATE_TX.validator,
                    value: testData.TEST_DELEGATE_TX.sendMessage.value.amount.amount,
                    coin: 'tzeta',
                },
            ]);
        });
        it('should build a undelegate txn from raw signed base64', function () {
            tx.enrichTransactionDetailsFromRawTransaction(testData.TEST_UNDELEGATE_TX.signedTxBase64);
            const json = tx.toJson();
            should_1.default.equal(json.sequence, testData.TEST_UNDELEGATE_TX.sequence);
            should_1.default.deepEqual(json.gasBudget, testData.TEST_UNDELEGATE_TX.gasBudget);
            should_1.default.equal(Buffer.from(json.publicKey, 'hex').toString('base64'), testData.TEST_UNDELEGATE_TX.pubKey);
            should_1.default.equal(json.sendMessages[0].value.validatorAddress, testData.TEST_UNDELEGATE_TX.sendMessage.value.validatorAddress);
            should_1.default.deepEqual(json.sendMessages[0].value.amount, testData.TEST_UNDELEGATE_TX.sendMessage.value.amount);
            should_1.default.equal(Buffer.from(json.signature).toString('base64'), testData.TEST_UNDELEGATE_TX.signature);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingDeactivate);
            tx.loadInputsAndOutputs();
            should_1.default.deepEqual(tx.inputs, [
                {
                    address: testData.TEST_UNDELEGATE_TX.delegator,
                    value: testData.TEST_UNDELEGATE_TX.sendMessage.value.amount.amount,
                    coin: 'tzeta',
                },
            ]);
            should_1.default.deepEqual(tx.outputs, [
                {
                    address: testData.TEST_UNDELEGATE_TX.validator,
                    value: testData.TEST_UNDELEGATE_TX.sendMessage.value.amount.amount,
                    coin: 'tzeta',
                },
            ]);
        });
        it('should build a redelegate txn from raw signed base64', function () {
            tx.enrichTransactionDetailsFromRawTransaction(testData.TEST_REDELEGATE_TX.signedTxBase64);
            const json = tx.toJson();
            should_1.default.equal(json.sequence, testData.TEST_REDELEGATE_TX.sequence);
            should_1.default.deepEqual(json.gasBudget, testData.TEST_REDELEGATE_TX.gasBudget);
            should_1.default.equal(Buffer.from(json.publicKey, 'hex').toString('base64'), testData.TEST_REDELEGATE_TX.pubKey);
            should_1.default.equal(json.sendMessages[0].value.validatorSrcAddress, testData.TEST_REDELEGATE_TX.sendMessage.value.validatorSrcAddress);
            should_1.default.deepEqual(json.sendMessages[0].value.amount, testData.TEST_REDELEGATE_TX.sendMessage.value.amount);
            should_1.default.equal(Buffer.from(json.signature).toString('base64'), testData.TEST_REDELEGATE_TX.signature);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingRedelegate);
            tx.loadInputsAndOutputs();
            should_1.default.deepEqual(tx.inputs, [
                {
                    address: testData.TEST_REDELEGATE_TX.delegator,
                    value: testData.TEST_REDELEGATE_TX.sendMessage.value.amount.amount,
                    coin: 'tzeta',
                },
            ]);
            should_1.default.deepEqual(tx.outputs, [
                {
                    address: testData.TEST_REDELEGATE_TX.validator,
                    value: testData.TEST_REDELEGATE_TX.sendMessage.value.amount.amount,
                    coin: 'tzeta',
                },
            ]);
        });
        it('should build a withdraw rewards from raw signed base64', function () {
            tx.enrichTransactionDetailsFromRawTransaction(testData.TEST_WITHDRAW_REWARDS_TX.signedTxBase64);
            const json = tx.toJson();
            should_1.default.equal(json.sequence, testData.TEST_WITHDRAW_REWARDS_TX.sequence);
            should_1.default.deepEqual(json.gasBudget, testData.TEST_WITHDRAW_REWARDS_TX.gasBudget);
            should_1.default.equal(Buffer.from(json.publicKey, 'hex').toString('base64'), testData.TEST_WITHDRAW_REWARDS_TX.pubKey);
            should_1.default.equal(json.sendMessages[0].value.validatorAddress, testData.TEST_WITHDRAW_REWARDS_TX.sendMessage.value.validatorAddress);
            should_1.default.equal(Buffer.from(json.signature).toString('base64'), testData.TEST_WITHDRAW_REWARDS_TX.signature);
            should_1.default.equal(tx.type, sdk_core_1.TransactionType.StakingWithdraw);
            tx.loadInputsAndOutputs();
            should_1.default.deepEqual(tx.inputs, [
                {
                    address: testData.TEST_WITHDRAW_REWARDS_TX.delegator,
                    value: 'UNAVAILABLE',
                    coin: 'tzeta',
                },
            ]);
            should_1.default.deepEqual(tx.outputs, [
                {
                    address: testData.TEST_WITHDRAW_REWARDS_TX.validator,
                    value: 'UNAVAILABLE',
                    coin: 'tzeta',
                },
            ]);
        });
        it('should fail to build a transfer from incorrect raw hex', function () {
            should_1.default.throws(() => tx.enrichTransactionDetailsFromRawTransaction('random' + testData.TEST_SEND_TX.signedTxBase64), 'incorrect raw data');
        });
        it('should fail to explain transaction with invalid raw hex', function () {
            should_1.default.throws(() => tx.enrichTransactionDetailsFromRawTransaction('randomString'), 'Invalid transaction');
        });
    });
    describe('Explain transaction', () => {
        it('should explain a transfer pay transaction', function () {
            tx.enrichTransactionDetailsFromRawTransaction(testData.TEST_SEND_TX.signedTxBase64);
            const explainedTransaction = tx.explainTransaction();
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: testData.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: testData.TEST_SEND_TX.recipient,
                        amount: testData.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: testData.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: testData.TEST_SEND_TX.feeAmount },
                type: 0,
            });
        });
        it('should fail to explain transaction with invalid raw base64 string', function () {
            should_1.default.throws(() => tx.enrichTransactionDetailsFromRawTransaction('randomString'), 'Invalid transaction');
        });
    });
});
