// @bitgo/sdk-coin-ada/dist/test/unit/messages/cip8/cip8Message.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const sinon_1 = __importDefault(require("sinon"));
const buffer_1 = require("buffer");
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../../../src");
const cip8Resources_1 = require("../../../resources/cip8Resources");
describe('Cip8Message', function () {
    let sandbox;
    beforeEach(function () {
        sandbox = sinon_1.default.createSandbox();
    });
    afterEach(function () {
        sandbox.restore();
    });
    const createDefaultMessageOptions = (message = cip8Resources_1.cip8TestResources.messages.simple) => {
        return {
            coinConfig: {},
            payload: message,
            signers: [cip8Resources_1.cip8TestResources.address.bech32],
        };
    };
    describe('constructor', function () {
        it('should create a new instance with correct message type', function () {
            const message = new src_1.Cip8Message(createDefaultMessageOptions());
            should_1.default.exist(message);
            message.getType().should.equal(sdk_core_1.MessageStandardType.CIP8);
        });
        it('should initialize with provided options', function () {
            const options = createDefaultMessageOptions();
            options.metadata = { test: 'value' };
            const message = new src_1.Cip8Message(options);
            should_1.default.exist(message);
            should_1.default.exist(message.getPayload());
            message.getPayload().should.equal(options.payload);
            should_1.default.exist(message.getSigners());
            message.getSigners().should.deepEqual(options.signers);
            should_1.default.exist(message.getMetadata());
            should_1.default.deepEqual(message.getMetadata(), options.metadata);
        });
        it('should accept signatures in constructor', function () {
            const signatures = [
                {
                    signature: buffer_1.Buffer.from('test-signature'),
                    publicKey: { pub: cip8Resources_1.cip8TestResources.keyPair.pub },
                },
            ];
            const options = {
                ...createDefaultMessageOptions(),
                signatures,
            };
            const message = new src_1.Cip8Message(options);
            should_1.default.exist(message.getSignatures());
            message.getSignatures().should.deepEqual(signatures);
        });
    });
    describe('validateAndGetCommonSetup', function () {
        it('should throw error when payload is missing', function () {
            const options = createDefaultMessageOptions();
            options.payload = undefined;
            const message = new src_1.Cip8Message(options);
            should_1.default.throws(() => message.validateAndGetCommonSetup(), /Payload is required to build a CIP8 message/);
        });
        it('should throw error when signers are missing', function () {
            const options = createDefaultMessageOptions();
            options.signers = [];
            const message = new src_1.Cip8Message(options);
            should_1.default.throws(() => message.validateAndGetCommonSetup(), /A signer address is required to build a CIP8 message/);
        });
        it('should return address CBOR bytes for valid input', function () {
            const message = new src_1.Cip8Message(createDefaultMessageOptions());
            const result = message.validateAndGetCommonSetup();
            should_1.default.exist(result);
            should_1.default.exist(result.addressCborBytes);
        });
    });
    describe('getSignablePayload', function () {
        it('should return signable payload as buffer', async function () {
            const message = new src_1.Cip8Message(createDefaultMessageOptions());
            const payload = await message.getSignablePayload();
            should_1.default.exist(payload);
            should_1.default.ok(buffer_1.Buffer.isBuffer(payload));
        });
    });
    describe('getBroadcastableSignatures', function () {
        it('should return empty array when no signatures exist', function () {
            const message = new src_1.Cip8Message(createDefaultMessageOptions());
            const signatures = message.getBroadcastableSignatures();
            should_1.default.exist(signatures);
            signatures.should.be.an.Array();
            signatures.should.be.empty();
        });
        it('should convert signature to COSE format', function () {
            const testSignature = {
                signature: buffer_1.Buffer.from('test-signature'),
                publicKey: { pub: cip8Resources_1.cip8TestResources.keyPair.pub },
            };
            const options = {
                ...createDefaultMessageOptions(),
                signatures: [testSignature],
            };
            const message = new src_1.Cip8Message(options);
            const signatures = message.getBroadcastableSignatures();
            should_1.default.exist(signatures);
            signatures.should.be.an.Array();
            signatures.should.have.length(1);
            should_1.default.exist(signatures[0].signature);
            should_1.default.ok(buffer_1.Buffer.isBuffer(signatures[0].signature));
            should_1.default.exist(signatures[0].publicKey);
            should_1.default.equal(signatures[0].publicKey.pub, cip8Resources_1.cip8TestResources.keyPair.pub);
        });
        it('should throw when validation fails', function () {
            const testSignature = {
                signature: buffer_1.Buffer.from('test-signature'),
                publicKey: { pub: cip8Resources_1.cip8TestResources.keyPair.pub },
            };
            const options = {
                ...createDefaultMessageOptions(),
                payload: undefined,
                signatures: [testSignature],
            };
            const message = new src_1.Cip8Message(options);
            should_1.default.throws(() => message.getBroadcastableSignatures(), /Payload is required to build a CIP8 message/);
        });
    });
});
