// @bitgo/public-types/src/schema/webhook/organizationWebhooks.ts
import { OrganizationWebhookType } from "./webhookType";
import { NonEmptyString } from "io-ts-types";
import { Optional, optionalString, optionalStringOrNumber } from "../../utils";
import * as t from "io-ts";
import { Webhook } from "./webhook";

export const AddOrganizationWebhookRequestBody = {
  type: OrganizationWebhookType,
  url: NonEmptyString,
  label: Optional(t.string),
};
export const AddOrganizationWebhookRequestBodyC = t.type({
  ...AddOrganizationWebhookRequestBody,
});

export type AddOrganizationWebhookRequestBody = t.TypeOf<
  typeof AddOrganizationWebhookRequestBodyC
>;

export const ListOrganizationWebhooksRequestBody = {
  type: Optional(OrganizationWebhookType),
  limit: optionalStringOrNumber,
  prevId: optionalString,
};

export const ListOrganizationWebhooksRequestBodyC = t.type({
  ...ListOrganizationWebhooksRequestBody,
});

export type ListOrganizationWebhooksRequestBody = t.TypeOf<
  typeof ListOrganizationWebhooksRequestBodyC
>;

export const ListOrganizationWebhooksResponse = t.intersection([
  t.type({
    webhooks: t.array(Webhook),
  }),
  t.partial({
    nextBatchPrevId: Optional(t.union([t.string, t.undefined])),
  }),
]);
export type ListOrganizationWebhooksResponse = t.TypeOf<
  typeof ListOrganizationWebhooksResponse
>;
