// @bitgo/public-types/src/schema/webhook/webhookType.ts
import * as t from "io-ts";

export enum WebhookTypeEnum {
  // Webhook for txRequest.state changes
  "txRequest" = "txRequest",
  // Webhook for any txRequest.transactions[].state changes
  "txRequestTransaction" = "txRequestTransaction",
  "transfer" = "transfer",
  "transaction" = "transaction",
  "transactionRemoved" = "transactionRemoved",
  "transactionExpire" = "transactionExpire",
  "pendingapproval" = "pendingapproval",
  "block" = "block",
  "admin" = "admin",
  "address_confirmation" = "address_confirmation",
  "lowFee" = "lowFee",
  "circuitBreaker" = "circuitBreaker",

  "lowFeeAddressBalance" = "lowFeeAddressBalance",

  // Specific to User Webhooks
  "wallet_confirmation" = "wallet_confirmation",

  "bankAccount" = "bankAccount",
  "userKycState" = "userKycState",
  "enterpriseKycState" = "enterpriseKycState",
  "identityStatus" = "identityStatus",

  "accessToken" = "accessToken",
  "accountCreated" = "accountCreated",
  "contractSigned" = "contractSigned",
  "fundsDeposited" = "fundsDeposited",
  "kycResult" = "kycResult",

  "policyBalanceChange" = "policyBalanceChange",

  // Webhook for reversal events (ACH reversals, fiat transaction reversals)
  "reversal" = "reversal",

  // Webhook for l2 transaction finality on l1
  "transaction_finality_on_l1" = "transaction_finality_on_l1",

  // Webhook for stuck transaction notification
  "stuckTx" = "stuckTx",

  // Webhook for trade order events
  "tradeOrder" = "tradeOrder",
}

export const WebhookType = t.keyof(WebhookTypeEnum, "WebhookType");
export type WebhookType = t.TypeOf<typeof WebhookType>;

export enum WalletWebhookTypeEnum {
  "txRequest" = "txRequest",
  "txRequestTransaction" = "txRequestTransaction",
  "transfer" = "transfer",
  "transaction" = "transaction",
  "transactionRemoved" = "transactionRemoved",
  "transactionExpire" = "transactionExpire",
  "pendingapproval" = "pendingapproval",
  "block" = "block",
  "admin" = "admin",
  "address_confirmation" = "address_confirmation",
  "lowFee" = "lowFee",
  "circuitBreaker" = "circuitBreaker",
  "lowFeeAddressBalance" = "lowFeeAddressBalance",
  "transaction_finality_on_l1" = "transaction_finality_on_l1",
  "fundsDeposited" = "fundsDeposited",
  "stuckTx" = "stuckTx",
}

export const WalletWebhookType = t.keyof(
  WalletWebhookTypeEnum,
  "WalletWebhookType",
);
export type WalletWebhookType = t.TypeOf<typeof WalletWebhookType>;

export enum EnterpriseWebhookTypeEnum {
  "bankAccount" = "bankAccount",
  "accessToken" = "accessToken",
  "policyBalanceChange" = "policyBalanceChange",
  "reversal" = "reversal",
}
export const EnterpriseWebhookType = t.keyof(
  EnterpriseWebhookTypeEnum,
  "EnterpriseWebhookType",
);
export type EnterpriseWebhookType = t.TypeOf<typeof EnterpriseWebhookType>;

export enum OrganizationWebhookTypeEnum {
  "transfer" = "transfer",
  "txRequest" = "txRequest",
  "pendingapproval" = "pendingapproval",
  "txRequestTransaction" = "txRequestTransaction",
  "bankAccount" = "bankAccount",
  "identityStatus" = "identityStatus",
  "userKycState" = "userKycState",
  "enterpriseKycState" = "enterpriseKycState",
  "accessToken" = "accessToken",
  "reversal" = "reversal",
  "tradeOrder" = "tradeOrder",
}
export const OrganizationWebhookType = t.keyof(
  OrganizationWebhookTypeEnum,
  "OrganizationWebhookType",
);
export type OrganizationWebhookType = t.TypeOf<typeof OrganizationWebhookType>;
