// @bitgo/sdk-coin-iota/dist/src/lib/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const transactionBuilder_1 = require("./transactionBuilder");
const transferTransaction_1 = require("./transferTransaction");
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
/**
 * Builder for IOTA transfer transactions.
 * Handles transactions that transfer IOTA tokens from the sender to one or more recipients.
 */
class TransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new transferTransaction_1.TransferTransaction(_coinConfig);
    }
    /**
     * Sets the recipients for this transfer transaction.
     * @param recipients - Array of recipients with their addresses and amounts
     * @returns This builder for method chaining
     * @throws BuildTransactionError if any recipient is invalid
     */
    recipients(recipients) {
        this.validateRecipients(recipients);
        this.transferTransaction.recipients = recipients;
        return this;
    }
    /**
     * Sets the payment objects (coin objects) to be used for this transfer.
     * These are the source coins that will be split and transferred to recipients.
     * @param paymentObjects - Array of IOTA coin objects to use for payment
     * @returns This builder for method chaining
     */
    paymentObjects(paymentObjects) {
        this.transferTransaction.paymentObjects = paymentObjects;
        return this;
    }
    /**
     * Initializes the builder with data from an existing transfer transaction.
     * @param tx - The source transaction to copy data from
     * @throws BuildTransactionError if transaction is not a TransferTransaction
     */
    initBuilder(tx) {
        this.validateTransactionType(tx);
        super.initBuilder(tx);
        this.copyTransferData(tx);
    }
    /**
     * Validates that the transaction is of the correct type (TransferTransaction).
     */
    validateTransactionType(tx) {
        if (!(tx instanceof transferTransaction_1.TransferTransaction)) {
            throw new sdk_core_1.BuildTransactionError('Transaction must be of type TransferTransaction');
        }
    }
    /**
     * Copies transfer-specific data from source transaction.
     */
    copyTransferData(tx) {
        this.transferTransaction.recipients = tx.recipients;
        this.transferTransaction.paymentObjects = tx.paymentObjects;
    }
    /**
     * Validates a complete transfer transaction.
     * @param transaction - The transaction to validate
     * @throws BuildTransactionError if transaction is invalid or not a TransferTransaction
     */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new sdk_core_1.BuildTransactionError('Transaction is required for validation');
        }
        this.validateTransactionType(transaction);
        super.validateTransaction(transaction);
        this.validateRecipients(transaction.recipients);
    }
    /**
     * Returns the transaction as a TransferTransaction type.
     */
    get transferTransaction() {
        return this._transaction;
    }
    /**
     * Validates all recipients have valid addresses and amounts.
     * @param recipients - Array of recipients to validate
     * @throws BuildTransactionError if any recipient has invalid address or amount
     */
    validateRecipients(recipients) {
        recipients.forEach((recipient) => {
            this.validateAddress({ address: recipient.address });
            this.validateValue(new bignumber_js_1.default(recipient.amount));
        });
    }
}
exports.TransferBuilder = TransferBuilder;
