// @bitgo/sdk-coin-iota/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transferBuilder_1 = require("./transferBuilder");
const constants_1 = require("./constants");
const utils_1 = __importDefault(require("./utils"));
const transactions_1 = require("@iota/iota-sdk/transactions");
/**
 * Factory class for creating IOTA transaction builders.
 *
 * This factory provides methods to create transaction builders for different
 * transaction types and to reconstruct transactions from raw transaction data.
 *
 * @example
 * ```typescript
 * const factory = new TransactionBuilderFactory(coins.get('tiota'));
 *
 * // Create a new transfer builder
 * const builder = factory.getTransferBuilder();
 *
 * // Rebuild from raw transaction
 * const rebuiltBuilder = factory.from(rawTxHex);
 * ```
 */
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(coinConfig) {
        super(coinConfig);
    }
    /**
     * Wallet initialization is not implemented for IOTA.
     * IOTA wallets are initialized through the TSS flow.
     *
     * @throws Error always - not implemented
     */
    getWalletInitializationBuilder() {
        throw new Error('Method not implemented.');
    }
    /**
     * Creates a transfer transaction builder.
     * Optionally initializes the builder with data from an existing transaction.
     *
     * @param tx - Optional existing transaction to initialize the builder
     * @returns A new TransferBuilder instance
     *
     * @example
     * ```typescript
     * // Create a new transfer builder
     * const builder = factory.getTransferBuilder();
     *
     * // Initialize from existing transaction
     * const existingTx = await builder.build();
     * const newBuilder = factory.getTransferBuilder(existingTx);
     * ```
     */
    getTransferBuilder(tx) {
        return this.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    }
    /**
     * Reconstructs a transaction builder from raw transaction data.
     * Automatically identifies the transaction type and creates the appropriate builder.
     *
     * @param rawTx - Raw transaction data (hex string or Uint8Array)
     * @returns TransactionBuilder appropriate for the transaction type
     * @throws InvalidTransactionError if the transaction type is not supported
     *
     * @example
     * ```typescript
     * // From hex string
     * const builder = factory.from('0x1234...');
     *
     * // From Uint8Array
     * const builder = factory.from(new Uint8Array([...]));
     *
     * // Rebuild and access the transaction
     * const tx = await builder.build();
     * ```
     */
    from(rawTx) {
        const rawTxBase64 = utils_1.default.getBase64String(rawTx);
        const transactionType = this.identifyTransactionType(rawTxBase64);
        switch (transactionType) {
            case sdk_core_1.TransactionType.Send:
                return this.createTransferBuilderFromRawTx(rawTxBase64);
            default:
                throw new sdk_core_1.InvalidTransactionError(`Unsupported transaction type: ${transactionType}`);
        }
    }
    /**
     * Identifies the transaction type by analyzing its commands.
     * Currently supports transfer transactions (Send type).
     *
     * @param rawTx - Raw transaction in base64 format
     * @returns The identified transaction type
     * @throws InvalidTransactionError if transaction contains unsupported commands
     */
    identifyTransactionType(rawTx) {
        const txData = transactions_1.Transaction.from(rawTx).getData();
        if (this.isTransferTransaction(txData)) {
            return sdk_core_1.TransactionType.Send;
        }
        throw new sdk_core_1.InvalidTransactionError('Transaction contains unsupported commands');
    }
    /**
     * Checks if a transaction is a transfer transaction by validating its commands.
     * Transfer transactions only contain: SplitCoins, MergeCoins, and TransferObjects commands.
     *
     * @param txData - The parsed transaction data
     * @returns true if all commands are valid transfer commands
     */
    isTransferTransaction(txData) {
        return txData.commands.every((command) => constants_1.TRANSFER_TRANSACTION_COMMANDS.includes(command.$kind));
    }
    /**
     * Creates a TransferBuilder from raw transaction data.
     *
     * @param rawTxBase64 - Raw transaction in base64 format
     * @returns Initialized TransferBuilder
     */
    createTransferBuilderFromRawTx(rawTxBase64) {
        const builder = new transferBuilder_1.TransferBuilder(this._coinConfig);
        builder.fromImplementation(rawTxBase64);
        return builder;
    }
    /**
     * Initializes a builder with data from an existing transaction.
     * If no transaction is provided, returns the builder as-is.
     *
     * @param tx - Optional transaction to initialize from
     * @param builder - The builder to initialize
     * @returns The initialized builder
     */
    initializeBuilder(tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
