// @bitgo/sdk-coin-iota/dist/src/lib/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactions_1 = require("@iota/iota-sdk/transactions");
const utils_1 = require("@iota/iota-sdk/utils");
const blake2b_1 = __importDefault(require("@bitgo/blake2b"));
const constants_1 = require("./constants");
const utils_2 = __importDefault(require("./utils"));
/**
 * Base class for IOTA transactions.
 * Manages transaction state, gas data, signatures, and building/serialization.
 */
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(coinConfig) {
        super(coinConfig);
        this._sender = '';
        this._rebuildRequired = false;
        this._isSimulateTx = true;
    }
    // Gas budget getter/setter - marks rebuild required when changed
    get gasBudget() {
        return this._gasBudget;
    }
    set gasBudget(value) {
        this._gasBudget = value;
        this.markRebuildRequired();
    }
    // Gas payment objects getter/setter - marks rebuild required when changed
    get gasPaymentObjects() {
        return this._gasPaymentObjects;
    }
    set gasPaymentObjects(value) {
        this._gasPaymentObjects = value;
        this.markRebuildRequired();
    }
    // Gas price getter/setter - marks rebuild required when changed
    get gasPrice() {
        return this._gasPrice;
    }
    set gasPrice(value) {
        this._gasPrice = value;
        this.markRebuildRequired();
    }
    // Gas sponsor getter/setter - marks rebuild required when changed
    get gasSponsor() {
        return this._gasSponsor;
    }
    set gasSponsor(value) {
        this._gasSponsor = value;
        this.markRebuildRequired();
    }
    // Transaction sender getter/setter - marks rebuild required when changed
    get sender() {
        return this._sender;
    }
    set sender(value) {
        this._sender = value;
        this.markRebuildRequired();
    }
    /**
     * Indicates whether this is a simulate transaction (dry run) or a real transaction.
     * Simulate transactions use maximum gas values for estimation purposes.
     */
    get isSimulateTx() {
        return this._isSimulateTx;
    }
    set isSimulateTx(value) {
        if (!value) {
            this.validateTxDataForRealTransaction();
            this.markRebuildRequired();
        }
        this._isSimulateTx = value;
    }
    /**
     * Marks that the transaction needs to be rebuilt before it can be signed or broadcast.
     */
    markRebuildRequired() {
        this._rebuildRequired = true;
    }
    /**
     * Validates transaction data when switching from simulate to real transaction mode.
     */
    validateTxDataForRealTransaction() {
        try {
            this.validateTxData();
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Tx data validation failed: ${errorMessage}. {Cause: ${error}}`);
        }
    }
    /**
     * Returns the signable payload for this transaction.
     * This is the Blake2b hash of the transaction data with intent message.
     * @throws Error if transaction is in simulate mode or not built
     */
    get signablePayload() {
        if (this.isSimulateTx) {
            throw new Error('Cannot sign a simulate tx');
        }
        this.ensureTransactionIsBuilt();
        const intentMessage = this.messageWithIntent(this._txDataBytes);
        return Buffer.from((0, blake2b_1.default)(32).update(intentMessage).digest('binary'));
    }
    /**
     * Returns the transaction digest (ID).
     * @throws Error if transaction is not built or needs rebuilding
     */
    get id() {
        this.ensureTransactionIsBuilt();
        return transactions_1.TransactionDataBuilder.getDigestFromBytes(this._txDataBytes);
    }
    /**
     * Ensures the transaction is built and doesn't need rebuilding.
     * @throws Error if transaction is not built or rebuild is required
     */
    ensureTransactionIsBuilt() {
        if (this._txDataBytes === undefined || this._rebuildRequired) {
            throw new Error('Tx not built or a rebuild is required');
        }
    }
    /**
     * Adds a signature from the transaction sender.
     */
    addSignature(publicKey, signature) {
        this._signature = { publicKey, signature };
    }
    /**
     * Adds a signature from the gas sponsor (if different from sender).
     */
    addGasSponsorSignature(publicKey, signature) {
        this._gasSponsorSignature = { publicKey, signature };
    }
    /**
     * Checks if this transaction can be signed.
     * Only real transactions (not simulate) can be signed.
     */
    canSign(_key) {
        return !this.isSimulateTx;
    }
    /**
     * Returns the transaction fee (gas budget).
     */
    getFee() {
        return this.gasBudget?.toString();
    }
    get serializedGasSponsorSignature() {
        return this._serializedGasSponsorSignature;
    }
    get serializedSignature() {
        return this._serializedSignature;
    }
    /**
     * Serializes all signatures for the transaction.
     * Includes both sender signature and gas sponsor signature if present.
     */
    serializeSignatures() {
        this._signatures = [];
        if (this._signature) {
            this._serializedSignature = this.serializeSignature(this._signature);
            this._signatures.push(this._serializedSignature);
        }
        if (this._gasSponsorSignature) {
            this._serializedGasSponsorSignature = this.serializeSignature(this._gasSponsorSignature);
            this._signatures.push(this._serializedGasSponsorSignature);
        }
    }
    /**
     * Converts the transaction to broadcast format (base64 encoded).
     */
    async toBroadcastFormat() {
        const txDataBytes = await this.build();
        return (0, utils_1.toBase64)(txDataBytes);
    }
    /**
     * Builds the transaction bytes.
     * If in simulate mode, builds a dry run transaction with max gas values.
     * Otherwise, builds a real transaction with actual gas data.
     */
    async build() {
        if (this.isSimulateTx) {
            return this.buildDryRunTransaction();
        }
        return this.buildRealTransaction();
    }
    toJson() {
        return {
            sender: this.sender,
            gasBudget: this.gasBudget,
            gasPrice: this.gasPrice,
            gasPaymentObjects: this.gasPaymentObjects,
            gasSponsor: this.gasSponsor,
            type: this.type,
        };
    }
    parseFromJSON(txData) {
        this.sender = txData.sender;
        this.gasBudget = txData.gasBudget;
        this.gasPrice = txData.gasPrice;
        this.gasPaymentObjects = txData.gasPaymentObjects;
        if (txData.gasSponsor !== undefined) {
            this.gasSponsor = txData.gasSponsor;
        }
    }
    /**
     * Parses transaction data from its broadcast format (base64 or raw bytes).
     * Extracts sender, gas data, and gas sponsor information.
     */
    parseFromBroadcastTx(tx) {
        const txData = transactions_1.Transaction.from(tx).getData();
        this.parseSender(txData);
        this.parseGasData(txData);
    }
    /**
     * Parses the sender address from transaction data.
     */
    parseSender(txData) {
        if (txData.sender) {
            this.sender = txData.sender;
        }
    }
    /**
     * Parses gas-related data from transaction data.
     */
    parseGasData(txData) {
        const gasData = txData.gasData;
        if (!gasData) {
            this.gasBudget = undefined;
            this.gasPrice = undefined;
            this.gasPaymentObjects = undefined;
            this.gasSponsor = undefined;
            return;
        }
        this.gasBudget = gasData.budget ? Number(gasData.budget) : undefined;
        this.gasPrice = gasData.price ? Number(gasData.price) : undefined;
        this.gasPaymentObjects =
            gasData.payment && gasData.payment.length > 0
                ? gasData.payment.map((payment) => payment)
                : undefined;
        this.gasSponsor = gasData.owner || undefined;
    }
    /**
     * @inheritDoc
     */
    explainTransaction() {
        const result = this.toJson();
        const displayOrder = ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'];
        const outputs = [];
        const explanationResult = {
            displayOrder,
            id: this.id,
            outputs,
            outputAmount: '0',
            changeOutputs: [],
            changeAmount: '0',
            fee: { fee: this.gasBudget ? this.gasBudget.toString() : '' },
            sender: this.sender,
            sponsor: this.gasSponsor,
            type: this.type,
        };
        return this.explainTransactionImplementation(result, explanationResult);
    }
    /**
     * Updates the simulate transaction flag based on gas data availability.
     * If all gas data is present, switches to real transaction mode.
     */
    updateIsSimulateTx() {
        const hasAllGasData = this.gasBudget && this.gasPrice && this.gasPaymentObjects && this.gasPaymentObjects.length > 0;
        this.isSimulateTx = !hasAllGasData;
    }
    /**
     * Builds a dry run (simulate) transaction with maximum gas values.
     * Used for gas estimation without committing the transaction.
     */
    async buildDryRunTransaction() {
        this.validateTxDataImplementation();
        await this.populateTxData();
        const txDataBuilder = new transactions_1.TransactionDataBuilder(this._iotaTransaction.getData());
        return txDataBuilder.build({
            overrides: {
                gasData: {
                    budget: constants_1.MAX_GAS_BUDGET.toString(),
                    price: constants_1.MAX_GAS_PRICE.toString(),
                    payment: [],
                },
            },
        });
    }
    /**
     * Builds a real transaction with actual gas data.
     * Only builds if necessary (first time or rebuild required).
     */
    async buildRealTransaction() {
        if (this._txDataBytes === undefined || this._rebuildRequired) {
            this.validateTxData();
            await this.populateTxData();
            this.setGasDataOnTransaction();
            this._txDataBytes = await this._iotaTransaction.build();
            this._rebuildRequired = false;
        }
        this.serializeSignatures();
        return this._txDataBytes;
    }
    /**
     * Sets gas data on the IOTA transaction object.
     */
    setGasDataOnTransaction() {
        this._iotaTransaction.setGasPrice(this.gasPrice);
        this._iotaTransaction.setGasBudget(this.gasBudget);
        this._iotaTransaction.setGasPayment(this.gasPaymentObjects.slice(0, constants_1.MAX_GAS_PAYMENT_OBJECTS - 1));
    }
    /**
     * Populates the IOTA transaction with inputs, commands, and gas sponsor if applicable.
     */
    async populateTxData() {
        this._iotaTransaction = new transactions_1.Transaction();
        this.populateTxInputsAndCommands();
        // If gas sponsor is different from sender, set up sponsored transaction
        if (this.hasDifferentGasSponsor()) {
            await this.setupGasSponsoredTransaction();
        }
        this._iotaTransaction.setSender(this.sender);
    }
    /**
     * Checks if the transaction has a gas sponsor different from the sender.
     */
    hasDifferentGasSponsor() {
        return Boolean(this.gasSponsor && this._sender !== this.gasSponsor);
    }
    /**
     * Sets up a gas-sponsored transaction by building the transaction kind
     * and setting the gas owner.
     */
    async setupGasSponsoredTransaction() {
        const transactionKind = await this._iotaTransaction.build({ onlyTransactionKind: true });
        this._iotaTransaction = transactions_1.Transaction.fromKind(transactionKind);
        this._iotaTransaction.setGasOwner(this._gasSponsor);
    }
    /**
     * Serializes a signature into IOTA's expected format.
     * Format: [signature_scheme_flag (1 byte), signature, public_key]
     * Currently hardcoded to EDDSA (0x00) as IOTA only supports this scheme.
     */
    serializeSignature(signature) {
        const SIGNATURE_SCHEME_EDDSA = 0x00;
        const pubKey = Buffer.from(signature.publicKey.pub, 'hex');
        const serializedSignature = new Uint8Array(1 + signature.signature.length + pubKey.length);
        serializedSignature.set([SIGNATURE_SCHEME_EDDSA]);
        serializedSignature.set(signature.signature, 1);
        serializedSignature.set(pubKey, 1 + signature.signature.length);
        return (0, utils_1.toBase64)(serializedSignature);
    }
    /**
     * Validates all transaction data required for a real (non-simulate) transaction.
     */
    validateTxData() {
        this.validateTxDataImplementation();
        this.validateCommonTxData();
        this.validateSignatures();
    }
    /**
     * Validates common transaction data (sender, gas data).
     */
    validateCommonTxData() {
        if (!this.sender) {
            throw new sdk_core_1.InvalidTransactionError('Transaction sender is required');
        }
        if (!this.gasPrice) {
            throw new sdk_core_1.InvalidTransactionError('Gas price is required');
        }
        if (!this.gasBudget) {
            throw new sdk_core_1.InvalidTransactionError('Gas budget is required');
        }
        if (!this.gasPaymentObjects || this.gasPaymentObjects.length === 0) {
            throw new sdk_core_1.InvalidTransactionError('Gas payment objects are required');
        }
    }
    /**
     * Validates sender and gas sponsor signatures if present.
     */
    validateSignatures() {
        if (this._signature && !this.isValidSignature(this._signature)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid sender signature');
        }
        if (this._gasSponsorSignature && !this.isValidSignature(this._gasSponsorSignature)) {
            throw new sdk_core_1.InvalidTransactionError('Invalid gas sponsor signature');
        }
    }
    /**
     * Checks if a signature has valid public key and signature data.
     */
    isValidSignature(signature) {
        return utils_2.default.isValidPublicKey(signature.publicKey.pub) && utils_2.default.isValidSignature((0, utils_1.toBase64)(signature.signature));
    }
}
exports.Transaction = Transaction;
