// @bitgo/sdk-coin-iota/dist/src/lib/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const utils_1 = __importDefault(require("./utils"));
const utils_2 = require("@iota/iota-sdk/utils");
/**
 * Base class for IOTA transaction builders.
 * Provides common functionality for building and validating IOTA transactions.
 */
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(coinConfig) {
        super(coinConfig);
    }
    /**
     * Initializes the transaction builder with data from an existing transaction.
     * Copies sender, gas data, and gas sponsor information.
     */
    initBuilder(tx) {
        this.validateTransaction(tx);
        this.copyTransactionData(tx);
    }
    /**
     * Copies transaction data from the source transaction to the builder's transaction.
     */
    copyTransactionData(tx) {
        this.transaction.sender = tx.sender;
        this.transaction.gasPrice = tx.gasPrice;
        this.transaction.gasBudget = tx.gasBudget;
        this.transaction.gasPaymentObjects = tx.gasPaymentObjects;
        this.transaction.gasSponsor = tx.gasSponsor;
    }
    get transactionType() {
        return this.transaction.type;
    }
    get transaction() {
        return this._transaction;
    }
    set transaction(transaction) {
        this._transaction = transaction;
    }
    /**
     * Sets the sender address for this transaction.
     * @param senderAddress - The IOTA address that is sending this transaction
     * @returns This transaction builder for method chaining
     */
    sender(senderAddress) {
        this.validateAddress({ address: senderAddress });
        this.transaction.sender = senderAddress;
        return this;
    }
    /**
     * Sets the gas data for this transaction (budget, price, and payment objects).
     * @param gasData - The gas configuration including budget, price, and payment objects
     * @returns This transaction builder for method chaining
     */
    gasData(gasData) {
        this.validateGasData(gasData);
        this.setGasDataOnTransaction(gasData);
        return this;
    }
    /**
     * Sets gas data fields on the transaction.
     */
    setGasDataOnTransaction(gasData) {
        this.transaction.gasPrice = gasData.gasPrice;
        this.transaction.gasBudget = gasData.gasBudget;
        this.transaction.gasPaymentObjects = gasData.gasPaymentObjects;
    }
    /**
     * Sets the gas sponsor for this transaction.
     * The gas sponsor pays for transaction fees instead of the sender.
     * @param sponsorAddress - The IOTA address sponsoring this transaction's gas fees
     * @returns This transaction builder for method chaining
     */
    gasSponsor(sponsorAddress) {
        this.validateAddress({ address: sponsorAddress });
        this.transaction.gasSponsor = sponsorAddress;
        return this;
    }
    /**
     * Adds a signature from the transaction sender.
     * @param publicKey - The sender's public key
     * @param signature - The signature bytes
     * @throws BuildTransactionError if the signature or public key is invalid
     */
    addSignature(publicKey, signature) {
        this.validateSignatureData(publicKey, signature);
        this.transaction.addSignature(publicKey, signature);
    }
    /**
     * Adds a signature from the gas sponsor.
     * @param publicKey - The gas sponsor's public key
     * @param signature - The signature bytes
     * @throws BuildTransactionError if the signature or public key is invalid
     */
    addGasSponsorSignature(publicKey, signature) {
        this.validateSignatureData(publicKey, signature);
        this.transaction.addGasSponsorSignature(publicKey, signature);
    }
    /**
     * Validates that the signature and public key are in valid formats.
     */
    validateSignatureData(publicKey, signature) {
        if (!utils_1.default.isValidPublicKey(publicKey.pub) || !utils_1.default.isValidSignature((0, utils_2.toBase64)(signature))) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction signature');
        }
    }
    validateKey(key) {
        throw new Error('Method not implemented.');
    }
    /**
     * Validates an IOTA address format.
     * @throws BuildTransactionError if address is invalid
     */
    validateAddress(address, addressFormat) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address ' + address.address);
        }
    }
    /**
     * Validates that a numeric value is valid (not NaN and not negative).
     * @throws BuildTransactionError if value is invalid
     */
    validateValue(value) {
        if (value.isNaN()) {
            throw new sdk_core_1.BuildTransactionError('Invalid amount format');
        }
        if (value.isLessThan(0)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be less than zero');
        }
    }
    /**
     * Validates that a raw transaction string is properly formatted.
     * @throws ParseTransactionError if raw transaction is invalid
     */
    validateRawTransaction(rawTransaction) {
        if (!rawTransaction) {
            throw new sdk_core_1.ParseTransactionError('Invalid raw transaction: Undefined');
        }
        if (!utils_1.default.isValidRawTransaction(rawTransaction)) {
            throw new sdk_core_1.ParseTransactionError('Invalid raw transaction');
        }
    }
    /**
     * Validates a transaction object has all required fields.
     * @throws Error if transaction is undefined or has invalid data
     */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        this.validateAddress({ address: transaction.sender });
        this.validateGasData({
            gasBudget: transaction.gasBudget,
            gasPrice: transaction.gasPrice,
            gasPaymentObjects: transaction.gasPaymentObjects,
        });
        if (transaction.gasSponsor) {
            this.validateAddress({ address: transaction.gasSponsor });
        }
    }
    /**
     * Creates a transaction object from a raw transaction string or bytes.
     * @param rawTransaction - Raw transaction in base64 string or Uint8Array format
     * @returns The parsed transaction object
     * @throws BuildTransactionError if raw transaction is invalid
     */
    fromImplementation(rawTransaction) {
        if (!utils_1.default.isValidRawTransaction(rawTransaction)) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction');
        }
        this.transaction.parseFromBroadcastTx(rawTransaction);
        return this.transaction;
    }
    /**
     * Sign implementation - not supported for IOTA transactions.
     * IOTA transactions must be signed externally.
     */
    signImplementation(key) {
        throw new Error('Method not implemented.');
    }
    /**
     * Builds the transaction and prepares it for broadcast.
     * Automatically switches from simulate to real transaction mode if gas data is present.
     */
    async buildImplementation() {
        this.updateTransactionMode();
        await this.transaction.build();
        this.transaction.addInputsAndOutputs();
        return this.transaction;
    }
    /**
     * Updates the transaction mode based on gas data availability.
     * Switches to real transaction mode if all gas data is provided.
     */
    updateTransactionMode() {
        const hasCompleteGasData = this.transaction.gasPrice && this.transaction.gasBudget && this.transaction.gasPaymentObjects;
        if (hasCompleteGasData) {
            this.transaction.isSimulateTx = false;
        }
    }
    /**
     * Validates gas data values and presence.
     * @throws BuildTransactionError if gas data is invalid
     */
    validateGasData(gasData) {
        if (gasData.gasBudget) {
            this.validateValue(new bignumber_js_1.default(gasData.gasBudget));
        }
        if (gasData.gasPrice) {
            this.validateValue(new bignumber_js_1.default(gasData.gasPrice));
        }
        if (gasData.gasPaymentObjects && gasData.gasPaymentObjects.length === 0) {
            throw new sdk_core_1.BuildTransactionError('Gas input objects list is empty');
        }
    }
}
exports.TransactionBuilder = TransactionBuilder;
