// @bitgo/sdk-coin-iota/dist/src/lib/transferTransaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferTransaction = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transaction_1 = require("./transaction");
const transactions_1 = require("@iota/iota-sdk/transactions");
const utils_1 = require("@iota/iota-sdk/utils");
const cryptography_1 = require("@iota/iota-sdk/cryptography");
const bcs_1 = require("@iota/bcs");
const constants_1 = require("./constants");
const utils_2 = __importDefault(require("./utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class TransferTransaction extends transaction_1.Transaction {
    constructor(coinConfig) {
        super(coinConfig);
        this._type = sdk_core_1.TransactionType.Send;
        this._inputs = [];
        this._outputs = [];
    }
    get recipients() {
        return this._recipients;
    }
    set recipients(value) {
        this._recipients = value;
        this._rebuildRequired = true;
    }
    get paymentObjects() {
        return this._paymentObjects;
    }
    set paymentObjects(value) {
        this._paymentObjects = value;
        this._rebuildRequired = true;
    }
    /**
     * @inheritDoc
     */
    addInputsAndOutputs() {
        if (!this._iotaTransaction) {
            return;
        }
        const totalAmount = this.recipients.reduce((accumulator, current) => accumulator + Number(current.amount), 0);
        this._outputs = this.recipients.map((recipient, index) => ({
            address: recipient.address,
            value: recipient.amount.toString(),
            coin: this._coinConfig.name,
        }));
        this._inputs = [
            {
                address: this.sender,
                value: totalAmount.toString(),
                coin: this._coinConfig.name,
            },
        ];
    }
    toJson() {
        return {
            ...super.toJson(),
            recipients: this.recipients,
            paymentObjects: this.paymentObjects,
        };
    }
    parseFromJSON(txData) {
        super.parseFromJSON(txData);
        this.recipients = txData.recipients;
        this.paymentObjects = txData.paymentObjects;
        this.updateIsSimulateTx();
    }
    /**
     * Parses a transfer transaction from its broadcast format (base64 or raw bytes).
     * Extracts recipients, amounts, and payment objects from the transaction data.
     */
    parseFromBroadcastTx(tx) {
        const txData = transactions_1.Transaction.from(tx).getData();
        this.validateTransferCommands(txData);
        super.parseFromBroadcastTx(tx);
        const { inputObjects, amounts, receivers } = this.parseTransactionInputs(txData);
        this.validateAmountsMatchReceivers(amounts, receivers);
        this.assignParsedObjects(inputObjects);
        this.assignRecipients(receivers, amounts);
        this.updateIsSimulateTx();
    }
    /**
     * Validates that the transaction only contains supported transfer commands.
     */
    validateTransferCommands(txData) {
        const hasUnsupportedCommands = txData.commands.some((command) => !constants_1.TRANSFER_TRANSACTION_COMMANDS.includes(command.$kind));
        if (hasUnsupportedCommands) {
            throw new sdk_core_1.InvalidTransactionError('Unsupported commands in the transaction');
        }
    }
    /**
     * Parses transaction inputs to extract objects, amounts, and receiver addresses.
     */
    parseTransactionInputs(txData) {
        const inputObjects = [];
        const amounts = [];
        const receivers = [];
        txData.inputs.forEach((input) => {
            if (input.$kind === 'Object' && 'ImmOrOwnedObject' in input.Object) {
                inputObjects.push(input.Object.ImmOrOwnedObject);
            }
            if (input.$kind === 'Pure' && 'bytes' in input.Pure) {
                const value = (0, utils_1.fromBase64)(input.Pure.bytes);
                const hexValue = '0x' + (0, sdk_core_1.toHex)(value);
                if (utils_2.default.isValidAddress(hexValue)) {
                    receivers.push(hexValue);
                }
                else {
                    amounts.push(new bcs_1.BcsReader(value).read64().toString());
                }
            }
        });
        return { inputObjects, amounts, receivers };
    }
    /**
     * Validates that the number of amounts matches the number of receivers.
     */
    validateAmountsMatchReceivers(amounts, receivers) {
        if (amounts.length !== receivers.length) {
            throw new sdk_core_1.InvalidTransactionError('Count of amounts does not match count of receivers');
        }
    }
    /**
     * Assigns parsed input objects to either gas payment objects or payment objects.
     * If no gas objects exist and sender pays own gas, objects become gas objects.
     * Otherwise, they become payment objects.
     */
    assignParsedObjects(inputObjects) {
        if (inputObjects.length === 0) {
            return;
        }
        const noGasObjectsExist = !this.gasPaymentObjects || this.gasPaymentObjects.length === 0;
        const senderPaysOwnGas = !this.gasSponsor || this.sender === this.gasSponsor;
        if (noGasObjectsExist && senderPaysOwnGas) {
            this.gasPaymentObjects = inputObjects;
        }
        else {
            this._paymentObjects = inputObjects;
        }
    }
    /**
     * Creates and assigns recipients from parsed addresses and amounts.
     */
    assignRecipients(receivers, amounts) {
        this._recipients = receivers.map((address, index) => ({
            address,
            amount: amounts[index],
        }));
    }
    messageWithIntent(message) {
        return (0, cryptography_1.messageWithIntent)('TransactionData', message);
    }
    /**
     * Populates the IOTA transaction with inputs and commands for the transfer.
     * This determines which objects to use for payment (either payment objects or gas objects),
     * consolidates them into a single coin, and then splits/transfers to recipients.
     */
    populateTxInputsAndCommands() {
        const sourceCoin = this.getConsolidatedSourceCoin();
        this.splitAndTransferToRecipients(sourceCoin);
    }
    /**
     * Determines which objects to use as the payment source and consolidates them.
     * If payment objects are provided, use those. Otherwise, if the sender is paying
     * their own gas, use the gas objects for payment.
     */
    getConsolidatedSourceCoin() {
        if (this.hasPaymentObjects()) {
            return this.consolidatePaymentObjects();
        }
        return this.consolidateGasObjects();
    }
    /**
     * Checks if payment objects exist and if gas objects should be used instead.
     * Gas objects are used when: no payment objects exist AND sender pays their own gas.
     */
    hasPaymentObjects() {
        const hasPaymentObjects = this.paymentObjects && this.paymentObjects.length > 0;
        if (hasPaymentObjects) {
            return true;
        }
        // If no payment objects, only use gas objects if sender pays own gas
        const senderPaysOwnGas = !this.gasSponsor || this.gasSponsor === this.sender;
        return !senderPaysOwnGas;
    }
    /**
     * Consolidates payment objects into a single coin object.
     * If multiple payment objects exist, they are merged in batches.
     */
    consolidatePaymentObjects() {
        if (!this.paymentObjects || this.paymentObjects.length === 0) {
            throw new sdk_core_1.InvalidTransactionError('Payment objects are required');
        }
        const firstObject = this._iotaTransaction.object(transactions_1.Inputs.ObjectRef(this.paymentObjects[0]));
        // Single object doesn't need consolidation
        if (this.paymentObjects.length > 1) {
            // Merge extra objects into the first one
            this.mergeObjectsInBatches(firstObject, this.paymentObjects.slice(1), constants_1.MAX_INPUT_OBJECTS);
        }
        return firstObject;
    }
    /**
     * Consolidates gas payment objects into a single coin object.
     * If the number of gas objects exceeds the maximum, they are merged in batches.
     */
    consolidateGasObjects() {
        if (!this.gasPaymentObjects || this.gasPaymentObjects.length === 0) {
            throw new sdk_core_1.InvalidTransactionError('Gas payment objects are required');
        }
        const gasObject = this._iotaTransaction.gas;
        if (this.gasPaymentObjects.length > constants_1.MAX_GAS_PAYMENT_OBJECTS) {
            // Merge excess gas objects to stay within limits
            this.mergeObjectsInBatches(gasObject, this.gasPaymentObjects, constants_1.MAX_INPUT_OBJECTS);
        }
        return gasObject;
    }
    /**
     * Merges multiple coin objects into a target coin in batches.
     * This is necessary because IOTA has limits on the number of objects per merge command.
     *
     * @param targetCoin - The coin to merge into
     * @param objectsToMerge - Array of objects to merge
     * @param batchSize - Maximum number of objects to merge per batch
     * @returns The consolidated coin object
     */
    mergeObjectsInBatches(targetCoin, objectsToMerge, batchSize) {
        for (let startIndex = 0; startIndex < objectsToMerge.length; startIndex += batchSize) {
            const batch = objectsToMerge.slice(startIndex, startIndex + batchSize);
            const batchAsTransactionObjects = batch.map((obj) => this._iotaTransaction.object(transactions_1.Inputs.ObjectRef(obj)));
            this._iotaTransaction.mergeCoins(targetCoin, batchAsTransactionObjects);
        }
    }
    /**
     * Splits the source coin into the amounts needed for each recipient and transfers them.
     * This creates split coin commands for all recipient amounts, then transfer commands
     * to send each split coin to the corresponding recipient.
     */
    splitAndTransferToRecipients(sourceCoin) {
        const recipientAmounts = this._recipients.map((recipient) => recipient.amount);
        const splitCoins = this._iotaTransaction.splitCoins(sourceCoin, recipientAmounts);
        this._recipients.forEach((recipient, index) => {
            this._iotaTransaction.transferObjects([splitCoins[index]], recipient.address);
        });
    }
    /**
     * Validates all transfer transaction data before building.
     * Checks recipients, payment objects, and ensures no duplicate object IDs.
     */
    validateTxDataImplementation() {
        this.validateRecipientsList();
        this.validatePaymentObjectsExist();
        this.validateNoDuplicateObjects();
    }
    /**
     * Validates that recipients exist and don't exceed the maximum allowed.
     */
    validateRecipientsList() {
        if (!this.recipients || this.recipients.length === 0) {
            throw new sdk_core_1.InvalidTransactionError('Transaction recipients are required');
        }
        if (this.recipients.length > constants_1.MAX_RECIPIENTS) {
            throw new sdk_core_1.InvalidTransactionError(`Recipients count (${this.recipients.length}) exceeds maximum allowed (${constants_1.MAX_RECIPIENTS})`);
        }
    }
    /**
     * Validates that either payment objects or gas objects exist for funding the transfer.
     * When a gas sponsor is used, payment objects are required.
     * Otherwise, either payment objects or gas objects can be used.
     */
    validatePaymentObjectsExist() {
        const hasPaymentObjects = this.paymentObjects && this.paymentObjects.length > 0;
        if (hasPaymentObjects) {
            return; // Payment objects exist, validation passes
        }
        // No payment objects - check if gas objects can be used instead
        const hasGasSponsor = this.gasSponsor && this.gasSponsor !== this.sender;
        if (hasGasSponsor) {
            throw new sdk_core_1.InvalidTransactionError('Payment objects are required when using a gas sponsor');
        }
        // No gas sponsor - gas objects must exist
        const hasGasObjects = this.gasPaymentObjects && this.gasPaymentObjects.length > 0;
        if (!hasGasObjects) {
            throw new sdk_core_1.InvalidTransactionError('Payment or Gas objects are required');
        }
    }
    /**
     * Validates that there are no duplicate object IDs within payment objects,
     * gas payment objects, or between the two groups.
     */
    validateNoDuplicateObjects() {
        const paymentObjectIds = this.paymentObjects?.map((obj) => obj.objectId) ?? [];
        // Check for duplicates within payment objects
        this.checkForDuplicateIds(paymentObjectIds, 'payment objects');
        if (!this.gasPaymentObjects || this.gasPaymentObjects.length === 0) {
            return;
        }
        const gasObjectIds = this.gasPaymentObjects.map((gas) => gas.objectId);
        // Check for duplicates within gas payment objects
        this.checkForDuplicateIds(gasObjectIds, 'gas payment objects');
        // Check for overlaps between payment and gas objects
        const overlappingIds = paymentObjectIds.filter((id) => gasObjectIds.includes(id));
        if (overlappingIds.length > 0) {
            throw new sdk_core_1.InvalidTransactionError('Payment objects cannot be the same as gas payment objects: ' + overlappingIds.join(', '));
        }
    }
    /**
     * Helper to check for duplicate IDs in an array.
     */
    checkForDuplicateIds(ids, objectType) {
        const uniqueIds = new Set(ids);
        if (uniqueIds.size !== ids.length) {
            throw new sdk_core_1.InvalidTransactionError(`Duplicate object IDs found in ${objectType}`);
        }
    }
    /**
     * @inheritDoc
     * Provides a human-readable explanation of the transfer transaction.
     */
    explainTransactionImplementation(_json, explanationResult) {
        const outputs = this.recipients.map((recipient) => recipient);
        const outputAmount = this.calculateTotalOutputAmount();
        return {
            ...explanationResult,
            outputAmount,
            outputs,
        };
    }
    /**
     * Calculates the total amount being transferred to all recipients.
     */
    calculateTotalOutputAmount() {
        return this.recipients
            .reduce((accumulator, current) => accumulator.plus(current.amount), new bignumber_js_1.default(0))
            .toString();
    }
}
exports.TransferTransaction = TransferTransaction;
