// @bitgo/secp256k1/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.bip32utils = exports.musig = exports.BIP32Factory = exports.bip32 = exports.ECPairFactory = exports.ECPair = exports.ecc = void 0;
const ecpair_1 = require("ecpair");
Object.defineProperty(exports, "ECPairFactory", { enumerable: true, get: function () { return ecpair_1.ECPairFactory; } });
const necc = __importStar(require("@noble/secp256k1"));
const bip32_1 = require("bip32");
Object.defineProperty(exports, "BIP32Factory", { enumerable: true, get: function () { return bip32_1.BIP32Factory; } });
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore base_crypto is exported as a subPath export, ignoring since compiler complains about importing like this
const baseCrypto = __importStar(require("@brandonblack/musig/base_crypto"));
const musig_1 = require("@brandonblack/musig");
const createHash = require('create-hash');
const createHmac = require('create-hmac');
necc.utils.sha256Sync = (...messages) => {
    const sha256 = createHash('sha256');
    for (const message of messages)
        sha256.update(message);
    return sha256.digest();
};
necc.utils.hmacSha256Sync = (key, ...messages) => {
    const hash = createHmac('sha256', Buffer.from(key));
    messages.forEach((m) => hash.update(m));
    return Uint8Array.from(hash.digest());
};
const defaultTrue = (param) => param !== false;
function throwToNull(fn) {
    try {
        return fn();
    }
    catch (e) {
        return null;
    }
}
function isPoint(p, xOnly) {
    if ((p.length === 32) !== xOnly)
        return false;
    try {
        return !!necc.Point.fromHex(p);
    }
    catch (e) {
        return false;
    }
}
function toBigInt(b) {
    const buff = Buffer.from(b);
    if (buff.length !== 32) {
        throw new Error('Invalid size ${buff.length}');
    }
    return BigInt(`0x${buff.toString('hex')}`);
}
const ecc = {
    isPoint: (p) => isPoint(p, false),
    isPrivate: (d) => necc.utils.isValidPrivateKey(d),
    isXOnlyPoint: (p) => isPoint(p, true),
    xOnlyPointAddTweak: (p, tweak) => throwToNull(() => {
        const P = necc.utils.pointAddScalar(p, tweak, true);
        const parity = P[0] % 2 === 1 ? 1 : 0;
        return { parity, xOnlyPubkey: P.slice(1) };
    }),
    pointFromScalar: (sk, compressed) => throwToNull(() => necc.getPublicKey(sk, defaultTrue(compressed))),
    pointCompress: (p, compressed) => {
        return necc.Point.fromHex(p).toRawBytes(defaultTrue(compressed));
    },
    pointMultiply: (a, tweak, compressed) => throwToNull(() => necc.utils.pointMultiply(a, tweak, defaultTrue(compressed))),
    pointAdd: (a, b, compressed) => throwToNull(() => {
        const A = necc.Point.fromHex(a);
        const B = necc.Point.fromHex(b);
        return A.add(B).toRawBytes(defaultTrue(compressed));
    }),
    pointAddScalar: (p, tweak, compressed) => throwToNull(() => necc.utils.pointAddScalar(p, tweak, defaultTrue(compressed))),
    privateAdd: (d, tweak) => throwToNull(() => {
        const res = necc.utils.privateAdd(d, tweak);
        // tiny-secp256k1 returns null rather than allowing a 0 private key to be returned
        // ECPair.testEcc() requires that behavior.
        if (res?.every((i) => i === 0))
            return null;
        return res;
    }),
    privateNegate: (d) => necc.utils.privateNegate(d),
    sign: (h, d, e) => {
        return necc.signSync(h, d, { der: false, extraEntropy: e });
    },
    signSchnorr: (h, d, e = Buffer.alloc(32, 0x00)) => {
        return necc.schnorr.signSync(h, d, e);
    },
    verify: (h, Q, signature, strict) => {
        return necc.verify(signature, h, Q, { strict });
    },
    recoverPublicKey: (h, signature, recovery, compressed) => {
        // Message hash must be exactly 32 bytes
        if (h.length !== 32) {
            return null;
        }
        // Signature must be exactly 64 bytes (r and s components)
        if (signature.length !== 64) {
            return null;
        }
        // Recovery value must be 0 or 1
        if (recovery !== 0 && recovery !== 1) {
            return null;
        }
        return throwToNull(() => necc.recoverPublicKey(h, signature, recovery, defaultTrue(compressed)));
    },
    verifySchnorr: (h, Q, signature) => {
        return necc.schnorr.verifySync(signature, h, Q);
    },
};
exports.ecc = ecc;
const crypto = {
    ...baseCrypto,
    pointMultiplyUnsafe(p, a, compress) {
        try {
            const product = necc.Point.fromHex(p).multiplyAndAddUnsafe(necc.Point.ZERO, toBigInt(a), BigInt(1));
            if (!product)
                return null;
            return product.toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointMultiplyAndAddUnsafe(p1, a, p2, compress) {
        try {
            const p2p = necc.Point.fromHex(p2);
            const p = necc.Point.fromHex(p1).multiplyAndAddUnsafe(p2p, toBigInt(a), BigInt(1));
            if (!p)
                return null;
            return p.toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointAdd(a, b, compress) {
        try {
            return necc.Point.fromHex(a).add(necc.Point.fromHex(b)).toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointAddTweak(p, tweak, compress) {
        try {
            const P = necc.Point.fromHex(p);
            const t = baseCrypto.readSecret(tweak);
            const Q = necc.Point.BASE.multiplyAndAddUnsafe(P, t, BigInt(1));
            if (!Q)
                throw new Error('Tweaked point at infinity');
            return Q.toRawBytes(compress);
        }
        catch {
            return null;
        }
    },
    pointCompress(p, compress = true) {
        return necc.Point.fromHex(p).toRawBytes(compress);
    },
    liftX(p) {
        try {
            return necc.Point.fromHex(p).toRawBytes(false);
        }
        catch {
            return null;
        }
    },
    getPublicKey(s, compress) {
        try {
            return necc.getPublicKey(s, compress);
        }
        catch {
            return null;
        }
    },
    taggedHash: necc.utils.taggedHashSync,
    sha256(...messages) {
        const sha256 = createHash('sha256');
        for (const message of messages)
            sha256.update(message);
        return sha256.digest();
    },
};
const ECPair = (0, ecpair_1.ECPairFactory)(ecc);
exports.ECPair = ECPair;
const bip32 = (0, bip32_1.BIP32Factory)(ecc);
exports.bip32 = bip32;
const musig = (0, musig_1.MuSigFactory)(crypto);
exports.musig = musig;
const bip32utils = __importStar(require("./bip32utils"));
exports.bip32utils = bip32utils;
