// @iota/iota-sdk/dist/cjs/utils/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  IOTA_CLOCK_OBJECT_ID: () => IOTA_CLOCK_OBJECT_ID,
  IOTA_DECIMALS: () => IOTA_DECIMALS,
  IOTA_FRAMEWORK_ADDRESS: () => IOTA_FRAMEWORK_ADDRESS,
  IOTA_SYSTEM_ADDRESS: () => IOTA_SYSTEM_ADDRESS,
  IOTA_SYSTEM_MODULE_NAME: () => IOTA_SYSTEM_MODULE_NAME,
  IOTA_SYSTEM_STATE_OBJECT_ID: () => IOTA_SYSTEM_STATE_OBJECT_ID,
  IOTA_TYPE_ARG: () => IOTA_TYPE_ARG,
  MOVE_STDLIB_ADDRESS: () => MOVE_STDLIB_ADDRESS,
  NANOS_PER_IOTA: () => NANOS_PER_IOTA
});
module.exports = __toCommonJS(constants_exports);
var import_iota_types = require("./iota-types.js");
const IOTA_DECIMALS = 9;
const NANOS_PER_IOTA = BigInt(1e9);
const MOVE_STDLIB_ADDRESS = "0x1";
const IOTA_FRAMEWORK_ADDRESS = "0x2";
const IOTA_SYSTEM_ADDRESS = "0x3";
const IOTA_CLOCK_OBJECT_ID = (0, import_iota_types.normalizeIotaObjectId)("0x6");
const IOTA_SYSTEM_MODULE_NAME = "iota_system";
const IOTA_TYPE_ARG = `${IOTA_FRAMEWORK_ADDRESS}::iota::IOTA`;
const IOTA_SYSTEM_STATE_OBJECT_ID = (0, import_iota_types.normalizeIotaObjectId)("0x5");
