// @iota/iota-sdk/dist/cjs/utils/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  CoinFormat: () => import_formatBalance.CoinFormat,
  IOTA_ADDRESS_LENGTH: () => import_iota_types.IOTA_ADDRESS_LENGTH,
  IOTA_CLOCK_OBJECT_ID: () => import_constants.IOTA_CLOCK_OBJECT_ID,
  IOTA_DECIMALS: () => import_constants.IOTA_DECIMALS,
  IOTA_FRAMEWORK_ADDRESS: () => import_constants.IOTA_FRAMEWORK_ADDRESS,
  IOTA_SYSTEM_ADDRESS: () => import_constants.IOTA_SYSTEM_ADDRESS,
  IOTA_SYSTEM_MODULE_NAME: () => import_constants.IOTA_SYSTEM_MODULE_NAME,
  IOTA_SYSTEM_STATE_OBJECT_ID: () => import_constants.IOTA_SYSTEM_STATE_OBJECT_ID,
  IOTA_TYPE_ARG: () => import_constants.IOTA_TYPE_ARG,
  MOVE_STDLIB_ADDRESS: () => import_constants.MOVE_STDLIB_ADDRESS,
  NANOS_PER_IOTA: () => import_constants.NANOS_PER_IOTA,
  deriveDynamicFieldID: () => import_dynamic_fields.deriveDynamicFieldID,
  formatAddress: () => import_format.formatAddress,
  formatAmount: () => import_formatBalance.formatAmount,
  formatBalance: () => import_formatBalance.formatBalance,
  formatDigest: () => import_format.formatDigest,
  formatType: () => import_format.formatType,
  formatWithSubscript: () => import_formatBalance.formatWithSubscript,
  fromB64: () => import_bcs.fromB64,
  fromBase58: () => import_bcs.fromBase58,
  fromBase64: () => import_bcs.fromBase64,
  fromHEX: () => import_bcs.fromHEX,
  fromHex: () => import_bcs.fromHex,
  isValidIotaAddress: () => import_iota_types.isValidIotaAddress,
  isValidIotaObjectId: () => import_iota_types.isValidIotaObjectId,
  isValidTransactionDigest: () => import_iota_types.isValidTransactionDigest,
  normalizeIotaAddress: () => import_iota_types.normalizeIotaAddress,
  normalizeIotaObjectId: () => import_iota_types.normalizeIotaObjectId,
  normalizeStructTag: () => import_iota_types.normalizeStructTag,
  parseAmount: () => import_parse.parseAmount,
  parseStructTag: () => import_iota_types.parseStructTag,
  safeParseAmount: () => import_parse.safeParseAmount,
  toB64: () => import_bcs.toB64,
  toBase58: () => import_bcs.toBase58,
  toBase64: () => import_bcs.toBase64,
  toHEX: () => import_bcs.toHEX,
  toHex: () => import_bcs.toHex,
  trimAddress: () => import_format.trimAddress,
  trimOrFormatAddress: () => import_format.trimOrFormatAddress
});
module.exports = __toCommonJS(utils_exports);
var import_format = require("./format.js");
var import_parse = require("./parse.js");
var import_formatBalance = require("./formatBalance.js");
var import_iota_types = require("./iota-types.js");
var import_bcs = require("@iota/bcs");
var import_constants = require("./constants.js");
var import_dynamic_fields = require("./dynamic-fields.js");
