// @bitgo/sdk-coin-world/dist/src/worldToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorldToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class WorldToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, WorldToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, WorldToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(WorldToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc **/
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /**
     * Make a query to World Chain explorer for information such as balance, token balance, solidity calls
     * @param {Object} query key-value pairs of parameters to append after /api
     * @returns {Promise<Object>} response World Chain explorer
     */
    async recoveryBlockchainExplorerQuery(query) {
        const apiToken = sdk_core_1.common.Environments[this.bitgo.getEnv()].worldExplorerApiToken;
        const explorerUrl = sdk_core_1.common.Environments[this.bitgo.getEnv()].worldExplorerBaseUrl;
        return await (0, abstract_eth_1.recoveryBlockchainExplorerQuery)(query, explorerUrl, apiToken);
    }
    getFullName() {
        return 'World Token';
    }
}
exports.WorldToken = WorldToken;
WorldToken.coinNames = {
    Mainnet: 'world',
    Testnet: 'tworld',
};
