// @bitgo/abstract-cosmos/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendMsgType = exports.ROOT_PATH = exports.UNAVAILABLE_TEXT = exports.groupVoteMsgTypeUrl = exports.groupProposalMsgTypeUrl = exports.executeContractMsgTypeUrl = exports.withdrawDelegatorRewardMsgTypeUrl = exports.redelegateTypeUrl = exports.undelegateMsgTypeUrl = exports.delegateMsgTypeUrl = exports.sendMsgTypeUrl = exports.DEFAULT_SEED_SIZE_BYTES = void 0;
exports.DEFAULT_SEED_SIZE_BYTES = 16;
exports.sendMsgTypeUrl = '/cosmos.bank.v1beta1.MsgSend';
exports.delegateMsgTypeUrl = '/cosmos.staking.v1beta1.MsgDelegate';
exports.undelegateMsgTypeUrl = '/cosmos.staking.v1beta1.MsgUndelegate';
exports.redelegateTypeUrl = '/cosmos.staking.v1beta1.MsgBeginRedelegate';
exports.withdrawDelegatorRewardMsgTypeUrl = '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward';
exports.executeContractMsgTypeUrl = '/cosmwasm.wasm.v1.MsgExecuteContract';
exports.groupProposalMsgTypeUrl = '/cosmos.group.v1.MsgSubmitProposal';
exports.groupVoteMsgTypeUrl = '/cosmos.group.v1.MsgVote';
exports.UNAVAILABLE_TEXT = 'UNAVAILABLE';
exports.ROOT_PATH = 'm/0';
exports.sendMsgType = '/types.MsgSend'; // thorchain uses this custom message type
