// @bitgo/sdk-coin-asi/dist/test/unit/asi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should = require("should");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sinon_1 = __importDefault(require("sinon"));
const sdk_test_1 = require("@bitgo/sdk-test");
const sdk_api_1 = require("@bitgo/sdk-api");
const index_1 = require("../../src/index");
const utils_1 = __importDefault(require("../../src/lib/utils"));
const asi_1 = require("../resources/asi");
const bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'test' });
const coinString = 'asi';
const testCoinString = 'tasi';
const coinFullName = 'Fetch Native';
const testCoinFullName = 'Testnet Fetch Native';
const baseFactor = 1e18;
const coinInstance = index_1.Asi.createInstance;
const testCoinInstance = index_1.Tasi.createInstance;
bitgo.safeRegister(testCoinString, testCoinInstance);
describe('Fetch Native', function () {
    let bitgo;
    let basecoin;
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, { env: 'mock' });
        bitgo.safeRegister(coinString, coinInstance);
        bitgo.safeRegister(testCoinString, testCoinInstance);
        bitgo.initializeTestVars();
        basecoin = bitgo.coin(testCoinString);
    });
    it('should return the right info', function () {
        const asi = bitgo.coin(coinString);
        const tasi = bitgo.coin(testCoinString);
        asi.getChain().should.equal(coinString);
        asi.getFamily().should.equal(coinString);
        asi.getFullName().should.equal(coinFullName);
        asi.getBaseFactor().should.equal(baseFactor);
        tasi.getChain().should.equal(testCoinString);
        tasi.getFamily().should.equal(coinString);
        tasi.getFullName().should.equal(testCoinFullName);
        tasi.getBaseFactor().should.equal(baseFactor);
    });
    describe('Address Validation', () => {
        it('should get address details without memoId', function () {
            const addressDetails = basecoin.getAddressDetails(asi_1.address.noMemoIdAddress);
            addressDetails.address.should.equal(asi_1.address.noMemoIdAddress);
            should.not.exist(addressDetails.memoId);
        });
        it('should get address details with memoId', function () {
            const addressDetails = basecoin.getAddressDetails(asi_1.address.validMemoIdAddress);
            addressDetails.address.should.equal(asi_1.address.validMemoIdAddress.split('?')[0]);
            addressDetails.memoId.should.equal('2');
        });
        it('should throw on invalid memo id address', () => {
            (() => {
                basecoin.getAddressDetails(asi_1.address.invalidMemoIdAddress);
            }).should.throw();
        });
        it('should throw on multiple memo id address', () => {
            (() => {
                basecoin.getAddressDetails(asi_1.address.multipleMemoIdAddress);
            }).should.throw();
        });
        it('should validate wallet receive address', async function () {
            const memo = 'memoId=7';
            const receiveAddress = {
                address: `${asi_1.TEST_ACCOUNT.pubAddress}?${memo}`,
                coinSpecific: {
                    rootAddress: asi_1.TEST_ACCOUNT.pubAddress,
                    memoID: '7',
                },
            };
            const isValid = await basecoin.isWalletAddress(receiveAddress);
            isValid.should.equal(true);
        });
        it('should validate account addresses correctly', () => {
            should.equal(utils_1.default.isValidAddress(asi_1.address.address1), true);
            should.equal(utils_1.default.isValidAddress(asi_1.address.address2), true);
            should.equal(utils_1.default.isValidAddress(asi_1.address.address3), false);
            should.equal(utils_1.default.isValidAddress(asi_1.address.address4), false);
            should.equal(utils_1.default.isValidAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidAddress(undefined), false);
            should.equal(utils_1.default.isValidAddress(''), false);
            should.equal(utils_1.default.isValidAddress(asi_1.address.validMemoIdAddress), true);
            should.equal(utils_1.default.isValidAddress(asi_1.address.invalidMemoIdAddress), false);
            should.equal(utils_1.default.isValidAddress(asi_1.address.multipleMemoIdAddress), false);
        });
        it('should validate validator addresses correctly', () => {
            should.equal(utils_1.default.isValidValidatorAddress(asi_1.address.validatorAddress1), true);
            should.equal(utils_1.default.isValidValidatorAddress(asi_1.address.validatorAddress2), true);
            should.equal(utils_1.default.isValidValidatorAddress(asi_1.address.validatorAddress3), false);
            should.equal(utils_1.default.isValidValidatorAddress(asi_1.address.validatorAddress4), false);
            should.equal(utils_1.default.isValidValidatorAddress('dfjk35y'), false);
            should.equal(utils_1.default.isValidValidatorAddress(undefined), false);
            should.equal(utils_1.default.isValidValidatorAddress(''), false);
        });
    });
    describe('Verify transaction: ', () => {
        it('should succeed to verify transaction', async function () {
            const txPrebuild = {
                txHex: asi_1.TEST_SEND_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: asi_1.TEST_SEND_TX.recipient,
                        amount: asi_1.TEST_SEND_TX.sendAmount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should succeed to verify sendMany transaction', async function () {
            const txPrebuild = {
                txHex: asi_1.TEST_SEND_MANY_TX.signedTxBase64,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: asi_1.TEST_SEND_MANY_TX.sendMessages[0].value.toAddress,
                        amount: asi_1.TEST_SEND_MANY_TX.sendMessages[0].value.amount[0].amount,
                    },
                    {
                        address: asi_1.TEST_SEND_MANY_TX.sendMessages[1].value.toAddress,
                        amount: asi_1.TEST_SEND_MANY_TX.sendMessages[1].value.amount[0].amount,
                    },
                ],
            };
            const verification = {};
            const isTransactionVerified = await basecoin.verifyTransaction({ txParams, txPrebuild, verification });
            isTransactionVerified.should.equal(true);
        });
        it('should fail to verify transaction with invalid param', async function () {
            const txPrebuild = {};
            const txParams = { recipients: undefined };
            await basecoin
                .verifyTransaction({
                txParams,
                txPrebuild,
            })
                .should.rejectedWith('missing required tx prebuild property txHex');
        });
    });
    describe('Explain Transaction: ', () => {
        it('should explain a transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: asi_1.TEST_SEND_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: asi_1.TEST_SEND_TX.hash,
                outputs: [
                    {
                        address: asi_1.TEST_SEND_TX.recipient,
                        amount: asi_1.TEST_SEND_TX.sendAmount,
                    },
                ],
                outputAmount: asi_1.TEST_SEND_TX.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: asi_1.TEST_SEND_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain sendMany transfer transaction', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: asi_1.TEST_SEND_MANY_TX.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: asi_1.TEST_SEND_MANY_TX.hash,
                outputs: [
                    {
                        address: asi_1.TEST_SEND_MANY_TX.sendMessages[0].value.toAddress,
                        amount: asi_1.TEST_SEND_MANY_TX.sendMessages[0].value.amount[0].amount,
                    },
                    {
                        address: asi_1.TEST_SEND_MANY_TX.sendMessages[1].value.toAddress,
                        amount: asi_1.TEST_SEND_MANY_TX.sendMessages[1].value.amount[0].amount,
                    },
                ],
                outputAmount: '20000',
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: asi_1.TEST_SEND_MANY_TX.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should explain a transfer transaction with memo', async function () {
            const explainedTransaction = await basecoin.explainTransaction({
                txHex: asi_1.TEST_TX_WITH_MEMO.signedTxBase64,
            });
            explainedTransaction.should.deepEqual({
                displayOrder: ['id', 'outputs', 'outputAmount', 'changeOutputs', 'changeAmount', 'fee', 'type'],
                id: asi_1.TEST_TX_WITH_MEMO.hash,
                outputs: [
                    {
                        address: asi_1.TEST_TX_WITH_MEMO.to,
                        amount: asi_1.TEST_TX_WITH_MEMO.sendAmount,
                        memo: asi_1.TEST_TX_WITH_MEMO.memo,
                    },
                ],
                outputAmount: asi_1.TEST_TX_WITH_MEMO.sendAmount,
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: asi_1.TEST_TX_WITH_MEMO.gasBudget.amount[0].amount },
                type: 0,
            });
        });
        it('should fail to explain transaction with missing params', async function () {
            try {
                await basecoin.explainTransaction({});
            }
            catch (error) {
                should.equal(error.message, 'missing required txHex parameter');
            }
        });
        it('should fail to explain transaction with invalid params', async function () {
            try {
                await basecoin.explainTransaction({ txHex: 'randomString' });
            }
            catch (error) {
                should.equal(error.message.startsWith('Invalid transaction:'), true);
            }
        });
    });
    describe('Parse Transactions: ', () => {
        const transferInputsResponse = {
            address: asi_1.TEST_SEND_TX.recipient,
            amount: new bignumber_js_1.default(asi_1.TEST_SEND_TX.sendAmount).plus(asi_1.TEST_SEND_TX.gasBudget.amount[0].amount).toFixed(),
        };
        const transferOutputsResponse = {
            address: asi_1.TEST_SEND_TX.recipient,
            amount: asi_1.TEST_SEND_TX.sendAmount,
        };
        it('should parse a transfer transaction', async function () {
            const parsedTransaction = await basecoin.parseTransaction({ txHex: asi_1.TEST_SEND_TX.signedTxBase64 });
            parsedTransaction.should.deepEqual({
                inputs: [transferInputsResponse],
                outputs: [transferOutputsResponse],
            });
        });
        it('should fail to parse a transfer transaction when explainTransaction response is undefined', async function () {
            const stub = sinon_1.default.stub(index_1.Asi.prototype, 'explainTransaction');
            stub.resolves(undefined);
            await basecoin
                .parseTransaction({ txHex: asi_1.TEST_SEND_TX.signedTxBase64 })
                .should.be.rejectedWith('Invalid transaction');
            stub.restore();
        });
    });
});
