// @bitgo/public-types/dist/src/schema/wallet/address.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressQueryResult = exports.WalletAddressUnion = exports.WalletAddress = exports.LightningWalletAddress = exports.BaseWalletAddress = exports.TrimmedBalance = exports.AddressChainString = exports.AddressChain = exports.AddressString = exports.AddressLabelString = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const coin_1 = require("../coin/coin");
exports.AddressLabelString = t.string;
exports.AddressString = t.string;
exports.AddressChain = t.union([
    t.literal(0),
    t.literal(1),
    t.literal(10),
    t.literal(11),
    t.literal(20),
    t.literal(21),
    t.literal(30),
    t.literal(31),
    t.literal(40),
    t.literal(41),
]);
exports.AddressChainString = t.union([
    t.literal("0"),
    t.literal("1"),
    t.literal("10"),
    t.literal("11"),
    t.literal("20"),
    t.literal("21"),
    t.literal("30"),
    t.literal("31"),
    t.literal("40"),
    t.literal("41"),
]);
exports.TrimmedBalance = t.intersection([
    t.type({
        balanceString: t.string,
        confirmedBalanceString: t.string,
        spendableBalanceString: t.string,
    }),
    t.partial({
        updated: t.union([io_ts_types_1.DateFromISOString, io_ts_types_1.date, t.unknown]),
        balance: t.number,
        lockedBalanceString: t.string,
        reservedBalanceString: t.string,
        totalReceived: t.number,
        totalSent: t.number,
        transferCount: t.number,
        tokens: t.UnknownRecord,
        nfts: t.UnknownRecord,
        unsupportedNfts: t.UnknownRecord,
        type: t.unknown,
        metadata: t.unknown,
    }),
]);
exports.BaseWalletAddress = t.partial({
    id: t.string,
    address: exports.AddressString,
    balance: t.union([exports.TrimmedBalance, t.unknown]),
    coinSpecific: t.unknown,
    label: exports.AddressLabelString,
    lastNonce: t.number,
    token: t.string,
    proof: t.string,
    signature: t.string,
    lastConsolidatedTime: io_ts_types_1.date,
    needsConsolidation: t.boolean,
    tokenConsolidationState: t.unknown,
    tokenAddress: t.unknown,
});
exports.LightningWalletAddress = t.intersection([
    exports.BaseWalletAddress,
    t.type({
        coin: coin_1.CoinString,
        wallet: t.string,
    }),
]);
exports.WalletAddress = t.intersection([
    exports.BaseWalletAddress,
    t.type({
        coin: coin_1.CoinString,
        wallet: t.string,
        chain: exports.AddressChain,
        index: t.number,
    }),
]);
exports.WalletAddressUnion = t.union([
    exports.WalletAddress,
    exports.LightningWalletAddress,
]);
exports.AddressQueryResult = t.intersection([
    t.type({
        coin: t.string,
    }),
    t.partial({
        pendingAddressCount: t.number,
        totalAddressCount: t.number,
        addresses: t.union([t.array(exports.WalletAddress), t.unknown]),
        nextBatchPrevId: t.string,
        count: t.number,
    }),
]);
