// @bitgo/account-lib/dist/test/unit/mpc/tss/ecdsa/ecdsa.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @prettier
 */
const assert_1 = __importDefault(require("assert"));
const crypto_1 = require("crypto");
const sdk_core_1 = require("@bitgo/sdk-core");
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const sinon = __importStar(require("sinon"));
const keccak_1 = __importDefault(require("keccak"));
const ecdsa_1 = require("../fixtures/ecdsa");
// Need to import and then monkey-patch to make stubbing work with ESM
const paillierBigintModule = require('paillier-bigint');
const paillierBigint = paillierBigintModule;
const schnorrModule = require('@bitgo/sdk-lib-mpc');
const Schnorr = schnorrModule.Schnorr;
describe('TSS ECDSA TESTS', function () {
    const MPC = new sdk_core_1.Ecdsa();
    const base = BigInt('0x010000000000000000000000000000000000000000000000000000000000000000'); // 2^256
    let keyShares;
    let commonPublicKey;
    const seed = Buffer.from('c4d1583a0b7b88626b56f0c83ee6df4d95d99cca73893ffb57c5e4411fa1b2b9c87456080e8d3f03462f065688abc28be2d4af3164d593c50b55269b435ea48d', 'hex');
    let A, B, C;
    before(async () => {
        // Direct monkey patching of the module for ESM compatibility
        const originalGenerateRandomKeys = paillierBigint.generateRandomKeys;
        const paillierMock = sinon.stub();
        paillierMock
            .onCall(0)
            .resolves(ecdsa_1.paillerKeys[0])
            .onCall(1)
            .resolves(ecdsa_1.paillerKeys[1])
            .onCall(2)
            .resolves(ecdsa_1.paillerKeys[2])
            .onCall(3)
            .resolves(ecdsa_1.paillerKeys[0])
            .onCall(4)
            .resolves(ecdsa_1.paillerKeys[1])
            .onCall(5)
            .resolves(ecdsa_1.paillerKeys[2]);
        // Replace the function directly on the module
        paillierBigint.generateRandomKeys = paillierMock;
        // Save references to restore later
        paillierMock.originalFn = originalGenerateRandomKeys;
        // Apply the same monkey-patching approach for Schnorr
        const originalCreateSchnorrProof = Schnorr.createSchnorrProof;
        const schnorrProofMock = sinon.stub();
        schnorrProofMock
            .onCall(0)
            .returns(ecdsa_1.schnorrProofs[0])
            .onCall(1)
            .returns(ecdsa_1.schnorrProofs[1])
            .onCall(2)
            .returns(ecdsa_1.schnorrProofs[2])
            .onCall(3)
            .returns(ecdsa_1.schnorrProofs[3])
            .onCall(4)
            .returns(ecdsa_1.schnorrProofs[4])
            .onCall(5)
            .returns(ecdsa_1.schnorrProofs[5]);
        // Replace the function directly on the module
        Schnorr.createSchnorrProof = schnorrProofMock;
        // Save references to restore later
        schnorrProofMock.originalFn = originalCreateSchnorrProof;
        [A, B, C] = await Promise.all([MPC.keyShare(1, 2, 3), MPC.keyShare(2, 2, 3), MPC.keyShare(3, 2, 3)]);
        // Needs to run this serially for testing deterministic key generation
        // to get specific paillier keys to be assigned
        const D = await MPC.keyShare(1, 2, 3, seed);
        const E = await MPC.keyShare(2, 2, 3, seed);
        const F = await MPC.keyShare(3, 2, 3, seed);
        const aKeyCombine = MPC.keyCombine(A.pShare, [B.nShares[1], C.nShares[1]]);
        const bKeyCombine = MPC.keyCombine(B.pShare, [A.nShares[2], C.nShares[2]]);
        const cKeyCombine = MPC.keyCombine(C.pShare, [A.nShares[3], B.nShares[3]]);
        // Shares with specific seeds
        const dKeyCombine = MPC.keyCombine(D.pShare, [E.nShares[1], F.nShares[1]]);
        const eKeyCombine = MPC.keyCombine(E.pShare, [D.nShares[2], F.nShares[2]]);
        const fKeyCombine = MPC.keyCombine(F.pShare, [D.nShares[3], E.nShares[3]]);
        // Shares for derived keys.
        const path = 'm/0/1';
        const aKeyDerive = MPC.keyDerive(A.pShare, [B.nShares[1], C.nShares[1]], path);
        const gKeyCombine = {
            xShare: aKeyDerive.xShare,
            yShares: aKeyCombine.yShares,
        };
        const hKeyCombine = MPC.keyCombine(B.pShare, [aKeyDerive.nShares[2], C.nShares[2]]);
        keyShares = [
            aKeyCombine,
            bKeyCombine,
            cKeyCombine,
            dKeyCombine,
            eKeyCombine,
            fKeyCombine,
            gKeyCombine,
            hKeyCombine,
        ];
        commonPublicKey = aKeyCombine.xShare.y;
        // Adding an explicit assertion to check if the stub was used
        paillierMock.callCount.should.equal(6, 'paillierMock should be called 6 times');
        schnorrProofMock.callCount.should.equal(8, 'schnorrProofMock should be called 6 times');
        // Restore original functions
        paillierBigint.generateRandomKeys = paillierMock.originalFn;
        Schnorr.createSchnorrProof = schnorrProofMock.originalFn;
    });
    describe('Ecdsa Key Generation Test', function () {
        it('should generate keys with correct threshold and share number', async function () {
            for (let index = 0; index < 3; index++) {
                const participantOne = (index % 3) + 1;
                const participantTwo = ((index + 1) % 3) + 1;
                const participantThree = ((index + 2) % 3) + 1;
                keyShares[index].xShare.i.should.equal(participantOne);
                keyShares[index].xShare.y.should.equal(commonPublicKey);
                keyShares[index].xShare.m.should.not.be.Null;
                keyShares[index].xShare.l.should.not.be.Null;
                keyShares[index].xShare.n.should.not.be.Null;
                const chaincode = BigInt('0x' + keyShares[index].xShare.chaincode);
                const isChainCodeValid = chaincode > BigInt(0) && chaincode <= base;
                isChainCodeValid.should.equal(true);
                keyShares[index].yShares[participantTwo].i.should.equal(participantOne);
                keyShares[index].yShares[participantThree].i.should.equal(participantOne);
                keyShares[index].yShares[participantTwo].j.should.equal(participantTwo);
                keyShares[index].yShares[participantThree].j.should.equal(participantThree);
                keyShares[index].yShares[participantTwo].n.should.not.be.Null;
                keyShares[index].yShares[participantThree].n.should.not.be.Null;
                const publicKeyPrefix = keyShares[index].xShare.y.slice(0, 2);
                const isRightPrefix = publicKeyPrefix === '03' || publicKeyPrefix === '02';
                isRightPrefix.should.equal(true);
            }
        });
        it('should generate keyshares with specific seed', async function () {
            // Keys should be deterministic when using seed
            const [, , , D, E, F] = keyShares;
            assert_1.default.deepEqual(D, ecdsa_1.mockDKeyShare);
            assert_1.default.deepEqual(E, ecdsa_1.mockEKeyShare);
            assert_1.default.deepEqual(F, ecdsa_1.mockFKeyShare);
        });
        it('should fail if seed is length less than 64 bytes', async function () {
            await MPC.keyShare(1, 2, 3, (0, crypto_1.randomBytes)(16)).should.be.rejectedWith('Seed must have a length of at least 64 bytes');
            await MPC.keyShare(1, 2, 3, (0, crypto_1.randomBytes)(32)).should.be.rejectedWith('Seed must have a length of at least 64 bytes');
        });
        it('should pass if seed length is greater than 64', async function () {
            const paillierMock = sinon.stub(paillierBigint, 'generateRandomKeys').onCall(0).resolves(ecdsa_1.paillerKeys[0]);
            const seed72Bytes = Buffer.from('4f7e914dc9ec696398675d1544aab61cb7a67662ffcbdb4079ec5d682be565d87c1b2de75c943dec14c96586984860268779498e6732473aed9ed9c2538f50bea0af926bdccc0134', 'hex');
            (await MPC.keyShare(1, 2, 3, seed72Bytes)).pShare.u.length.should.equal(64);
            paillierMock.restore();
        });
        it('should calculate correct chaincode while combining', async function () {
            const keyCombine = MPC.keyCombine(ecdsa_1.mockPShare, ecdsa_1.mockNShares);
            keyCombine.xShare.chaincode.should.equal('fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc32');
        });
        it('should fail to generate keys with invalid threshold and share number', async function () {
            const invalidConfigs = [
                { index: 1, threshold: 5, numShares: 3 },
                { index: -1, threshold: 2, numShares: 3 },
                { index: 1, threshold: 2, numShares: 1 },
            ];
            for (let index = 0; index < invalidConfigs.length; index++) {
                try {
                    await MPC.keyShare(invalidConfigs[index].index, invalidConfigs[index].threshold, invalidConfigs[index].numShares);
                }
                catch (e) {
                    e.should.equal('Invalid KeyShare Config');
                }
            }
        });
        it('should derive unhardened child keys', async function () {
            // parent key
            const aKeyCombine = keyShares[0];
            const commonKeychain = aKeyCombine.xShare.y + aKeyCombine.xShare.chaincode;
            for (let index = 0; index < 10; index++) {
                const path = `m/0/0/${index}`;
                const subkey = MPC.keyDerive(A.pShare, [B.nShares[1], C.nShares[1]], path);
                const derive1 = MPC.deriveUnhardened(commonKeychain, path);
                const derive2 = MPC.deriveUnhardened(commonKeychain, path);
                derive1.should.equal(derive2, 'derivation should be deterministic');
                (subkey.xShare.y + subkey.xShare.chaincode).should.equal(derive1, 'subkey common keychain should match derived keychain');
            }
        });
    });
    describe('ECDSA Signing', async function () {
        let config;
        before(() => {
            const [A, B, C, D, E, F, G, H] = keyShares;
            config = [
                { signerOne: A, signerTwo: B },
                { signerOne: A, signerTwo: C },
                { signerOne: B, signerTwo: A },
                { signerOne: B, signerTwo: C },
                { signerOne: C, signerTwo: A },
                { signerOne: C, signerTwo: B },
                // Checks signing with specific seed
                { signerOne: D, signerTwo: E },
                { signerOne: E, signerTwo: F },
                { signerOne: F, signerTwo: D },
                // Checks with specific hashing algorithm
                { signerOne: A, signerTwo: B, hash: 'keccak256' },
                // checks with no hashing
                { signerOne: A, signerTwo: B, shouldHash: false },
                // Checks with derived subkey
                { signerOne: G, signerTwo: H },
            ];
        });
        for (let index = 0; index < 9; index++) {
            it(`should properly sign the message case ${index}`, async function () {
                // Step One
                // signerOne, signerTwo have decided to sign the message
                const signerOne = config[index].signerOne;
                const signerOneIndex = signerOne.xShare.i;
                const signerTwo = config[index].signerTwo;
                const signerTwoIndex = signerTwo.xShare.i;
                const [signerOneToTwoPaillierChallenge, signerTwoToOnePaillierChallenge] = await Promise.all([
                    sdk_lib_mpc_1.EcdsaPaillierProof.generateP((0, sdk_core_1.hexToBigInt)(signerOne.yShares[signerTwoIndex].n)),
                    sdk_lib_mpc_1.EcdsaPaillierProof.generateP((0, sdk_core_1.hexToBigInt)(signerTwo.yShares[signerOneIndex].n)),
                ]);
                // Step Two
                // First signer generates their range proof challenge.
                const signerOneXShare = MPC.appendChallenge(signerOne.xShare, sdk_lib_mpc_1.EcdsaTypes.serializeNtilde(ecdsa_1.ntildes[index]), sdk_lib_mpc_1.EcdsaTypes.serializePaillierChallenge({ p: signerOneToTwoPaillierChallenge }));
                // Step Three
                //  Second signer generates their range proof challenge.
                const signerTwoXShare = MPC.appendChallenge(signerTwo.xShare, sdk_lib_mpc_1.EcdsaTypes.serializeNtilde(ecdsa_1.ntildes[index + 1]), sdk_lib_mpc_1.EcdsaTypes.serializePaillierChallenge({ p: signerTwoToOnePaillierChallenge }));
                const signerTwoChallenge = { ntilde: signerTwoXShare.ntilde, h1: signerTwoXShare.h1, h2: signerTwoXShare.h2 };
                // Step Four
                // First signer receives the challenge from the second signer and appends it to their YShare
                const signerTwoYShare = MPC.appendChallenge(signerOne.yShares[signerTwoIndex], signerTwoChallenge, sdk_lib_mpc_1.EcdsaTypes.serializePaillierChallenge({ p: signerTwoToOnePaillierChallenge }));
                // Step Five
                // Sign Shares are created by one of the participants (signerOne)
                // with its private XShare and YShare corresponding to the other participant (signerTwo)
                // This step produces a private WShare which signerOne saves and KShare which signerOne sends to signerTwo
                const signShares = await MPC.signShare(signerOneXShare, signerTwoYShare);
                // Step Six
                // signerTwo receives the KShare from signerOne and uses it produce private
                // BShare (Beta Share) which signerTwo saves and AShare (Alpha Share)
                // which is sent to signerOne
                const signConvertS21 = await MPC.signConvertStep1({
                    xShare: signerTwoXShare,
                    yShare: signerTwo.yShares[signerOneIndex], // YShare corresponding to the other participant signerOne
                    kShare: signShares.kShare,
                });
                // Step Seven
                // signerOne receives the AShare from signerTwo and signerOne using the private WShare from step two
                // uses it produce private GShare (Gamma Share) and MUShare (Mu Share) which
                // is sent to signerTwo to produce its Gamma Share
                const signConvertS12 = await MPC.signConvertStep2({
                    aShare: signConvertS21.aShare,
                    wShare: signShares.wShare,
                });
                // Step Eight
                // signerTwo receives the MUShare from signerOne and signerOne using the private BShare from step three
                // uses it produce private GShare (Gamma Share)
                const signConvertS21_2 = await MPC.signConvertStep3({
                    muShare: signConvertS12.muShare,
                    bShare: signConvertS21.bShare,
                });
                // Step Nine
                // signerOne and signerTwo both have successfully generated GShares and they use
                // the sign combine function to generate their private omicron shares and
                // delta shares which they share to each other
                const [signCombineOne, signCombineTwo] = [
                    MPC.signCombine({
                        gShare: signConvertS12.gShare,
                        signIndex: {
                            i: signConvertS12.muShare.i,
                            j: signConvertS12.muShare.j,
                        },
                    }),
                    MPC.signCombine({
                        gShare: signConvertS21_2.gShare,
                        signIndex: {
                            i: signConvertS21_2.signIndex.i,
                            j: signConvertS21_2.signIndex.j,
                        },
                    }),
                ];
                const MESSAGE = Buffer.from('TOO MANY SECRETS');
                // Step Ten
                // signerOne and signerTwo shares the delta share from each other
                // and finally signs the message using their private OShare
                // and delta share received from the other signer
                const hashGenerator = (hashType) => {
                    return hashType === 'keccak256' ? (0, keccak_1.default)('keccak256') : undefined;
                };
                const [signA, signB] = [
                    MPC.sign(MESSAGE, signCombineOne.oShare, signCombineTwo.dShare, hashGenerator(config[index].hash), config[index].shouldHash),
                    MPC.sign(MESSAGE, signCombineTwo.oShare, signCombineOne.dShare, hashGenerator(config[index].hash), config[index].shouldHash),
                ];
                // Step Eleven
                // Construct the final signature
                const signature = MPC.constructSignature([signA, signB]);
                // Step Twelve
                // Verify signature
                const isValid = MPC.verify(MESSAGE, signature, hashGenerator(config[index].hash), config[index].shouldHash);
                isValid.should.equal(true);
            });
        }
    });
});
