// @bitgo/abstract-utxo/dist/cjs/test/unit/transaction/fixedScript/explainPsbt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const strict_1 = __importDefault(require("node:assert/strict"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const fixedScript_1 = require("../../../../src/transaction/fixedScript");
const names_1 = require("../../../../src/names");
function describeTransactionWith(acidTest) {
    describe(`${acidTest.name}`, function () {
        let psbt;
        let psbtBytes;
        let walletXpubs;
        let customChangeWalletXpubs;
        let wasmPsbt;
        let refExplanation;
        before('prepare', function () {
            psbt = acidTest.createPsbt();
            const coinName = (0, names_1.getCoinName)(acidTest.network);
            refExplanation = (0, fixedScript_1.explainPsbt)(psbt, { pubs: acidTest.rootWalletKeys }, coinName, {
                strict: true,
            });
            psbtBytes = psbt.toBuffer();
            const networkName = utxolib.getNetworkName(acidTest.network);
            (0, strict_1.default)(networkName);
            walletXpubs = acidTest.rootWalletKeys.triple.map((k) => k.neutered().toBase58());
            customChangeWalletXpubs = acidTest.otherWalletKeys.triple.map((k) => k.neutered().toBase58());
            wasmPsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(psbtBytes, networkName);
        });
        it('should match the expected values for explainPsbt', function () {
            // note: `outputs` means external outputs here
            strict_1.default.strictEqual(refExplanation.outputs.length, 3);
            strict_1.default.strictEqual(refExplanation.changeOutputs.length, acidTest.outputs.length - 3);
            strict_1.default.strictEqual(refExplanation.outputAmount, '1800');
            strict_1.default.strictEqual(refExplanation.changeOutputs.length, acidTest.outputs.length - 3);
            refExplanation.changeOutputs.forEach((change) => {
                strict_1.default.strictEqual(change.amount, '900');
                strict_1.default.strictEqual(typeof change.address, 'string');
            });
        });
        it('reference implementation should support custom change outputs', function () {
            const coinName = (0, names_1.getCoinName)(acidTest.network);
            const customChangeExplanation = (0, fixedScript_1.explainPsbt)(psbt, { pubs: acidTest.rootWalletKeys, customChangePubs: acidTest.otherWalletKeys }, coinName, { strict: true });
            strict_1.default.ok(customChangeExplanation.customChangeOutputs);
            strict_1.default.strictEqual(customChangeExplanation.changeOutputs.length, refExplanation.changeOutputs.length);
            strict_1.default.strictEqual(customChangeExplanation.outputs.length, refExplanation.outputs.length - 1);
            strict_1.default.strictEqual(customChangeExplanation.customChangeOutputs.length, 1);
            strict_1.default.strictEqual(customChangeExplanation.customChangeOutputs[0].amount, '900');
        });
        it('should match explainPsbtWasm', function () {
            const wasmExplanation = (0, fixedScript_1.explainPsbtWasm)(wasmPsbt, walletXpubs, {
                replayProtection: {
                    publicKeys: [acidTest.getReplayProtectionPublicKey()],
                },
            });
            for (const key of Object.keys(refExplanation)) {
                const refValue = refExplanation[key];
                const wasmValue = wasmExplanation[key];
                switch (key) {
                    case 'displayOrder':
                    case 'inputSignatures':
                    case 'signatures':
                        // these are deprecated fields that we want to get rid of
                        strict_1.default.deepStrictEqual(wasmValue, undefined);
                        break;
                    default:
                        strict_1.default.deepStrictEqual(wasmValue, refValue, `mismatch for key ${key}`);
                        break;
                }
            }
        });
        if (acidTest.network !== utxolib.networks.bitcoin) {
            return;
        }
        // extended test suite for bitcoin
        it('returns custom change outputs when parameter is set', function () {
            const wasmExplanation = (0, fixedScript_1.explainPsbtWasm)(wasmPsbt, walletXpubs, {
                replayProtection: {
                    publicKeys: [acidTest.getReplayProtectionPublicKey()],
                },
                customChangeWalletXpubs,
            });
            strict_1.default.ok(wasmExplanation.customChangeOutputs);
            strict_1.default.deepStrictEqual(wasmExplanation.outputs.length, 2);
            strict_1.default.deepStrictEqual(wasmExplanation.customChangeOutputs.length, 1);
            strict_1.default.deepStrictEqual(wasmExplanation.customChangeOutputs[0].amount, '900');
        });
    });
}
describe('explainPsbt(Wasm)', function () {
    utxo_lib_1.testutil.AcidTest.suite().forEach((test) => describeTransactionWith(test));
});
