// @bitgo/abstract-utxo/dist/cjs/test/unit/transaction/fixedScript/parsePsbt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const strict_1 = __importDefault(require("node:assert/strict"));
const sinon = __importStar(require("sinon"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const sdk_core_1 = require("@bitgo/sdk-core");
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const parseTransaction_1 = require("../../../../src/transaction/fixedScript/parseTransaction");
const util_1 = require("../../util");
const fixedScript_1 = require("../../../../src/transaction/fixedScript");
const names_1 = require("../../../../src/names");
function getTxParamsFromExplanation(explanation, { externalCustomChangeAddress }) {
    // The external outputs are the ones that are in outputs but not in changeOutputs
    const changeAddresses = new Set(explanation.changeOutputs.map((o) => o.address));
    let externalOutputs = explanation.outputs.filter((o) => o.address && !changeAddresses.has(o.address));
    let changeAddress;
    if (externalCustomChangeAddress) {
        // convert an external output to a change output
        //
        // in combination with allowExternalChangeAddress, this allows an external
        // output on the transaction without a size constraint
        const externalOutput = externalOutputs[0];
        if (!externalOutput) {
            throw new Error('no external output found');
        }
        changeAddress = externalOutput.address;
        externalOutputs = externalOutputs.slice(1);
    }
    return {
        recipients: externalOutputs.map((output) => ({
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            address: output.address,
            amount: output.amount,
        })),
        changeAddress,
    };
}
function getChangeInfoFromPsbt(psbt) {
    try {
        return utxolib.bitgo.findInternalOutputIndices(psbt).map((i) => {
            const output = psbt.data.outputs[i];
            const derivations = output.bip32Derivation ?? output.tapBip32Derivation ?? undefined;
            if (!derivations || derivations.length !== 3) {
                throw new Error('expected 3 derivation paths');
            }
            const path = derivations[0].path;
            const { chain, index } = utxolib.bitgo.getChainAndIndexFromPath(path);
            return {
                address: wasm_utxo_1.address.fromOutputScriptWithCoin(psbt.txOutputs[i].script, (0, names_1.getCoinName)(psbt.network)),
                chain,
                index,
            };
        });
    }
    catch (e) {
        if (e instanceof utxolib.bitgo.ErrorNoMultiSigInputFound) {
            return undefined;
        }
        throw e;
    }
}
function describeParseTransactionWith(acidTest, label, backend, { txParams, externalCustomChangeAddress = false, expectedExplicitExternalSpendAmount, expectedImplicitExternalSpendAmount, txFormat = 'psbt', }) {
    describe(`${acidTest.name}/${label}`, function () {
        let refParsedTransaction;
        let coin;
        let mockWallet;
        let stubExplainTransaction;
        before('prepare', async function () {
            const coinName = (0, names_1.getCoinName)(acidTest.network);
            coin = (0, util_1.getUtxoCoin)(coinName);
            // Create PSBT and explanation
            const psbt = acidTest.createPsbt();
            const tx = psbt.getUnsignedTx();
            const txHash = tx.getId();
            let explanation;
            if (txFormat === 'psbt') {
                if (backend === 'utxolib') {
                    explanation = (0, fixedScript_1.explainPsbt)(psbt, { pubs: acidTest.rootWalletKeys }, coinName, {
                        strict: true,
                    });
                }
                else if (backend === 'wasm') {
                    const wasmPsbt = wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(psbt.toBuffer(), utxolib.getNetworkName(acidTest.network));
                    explanation = (0, fixedScript_1.explainPsbtWasm)(wasmPsbt, acidTest.rootWalletKeys.triple.map((k) => k.neutered().toBase58()), {
                        replayProtection: {
                            publicKeys: [acidTest.getReplayProtectionPublicKey()],
                        },
                    });
                }
                else {
                    throw new Error(`Invalid backend: ${backend}`);
                }
            }
            else if (txFormat === 'legacy') {
                const pubs = acidTest.rootWalletKeys.triple.map((k) => k.neutered().toBase58());
                // Extract change info from PSBT to pass to explainLegacyTx
                const changeInfo = getChangeInfoFromPsbt(psbt);
                explanation = (0, fixedScript_1.explainLegacyTx)(tx, { pubs, changeInfo }, coinName);
            }
            else {
                throw new Error(`Invalid txFormat: ${txFormat}`);
            }
            // Determine txParams
            let resolvedTxParams;
            if (txParams === 'inferFromExplanation' || txParams === undefined) {
                resolvedTxParams = getTxParamsFromExplanation(explanation, { externalCustomChangeAddress });
            }
            else if ('rbfTxIds' in txParams) {
                // Replace placeholder txHash with actual computed txHash
                resolvedTxParams = {
                    rbfTxIds: txParams.rbfTxIds.map((hash) => (hash === 'PLACEHOLDER' ? txHash : hash)),
                };
            }
            else {
                resolvedTxParams = txParams;
            }
            if (externalCustomChangeAddress) {
                resolvedTxParams.allowExternalChangeAddress = true;
            }
            // Create mock wallet
            mockWallet = sinon.createStubInstance(sdk_core_1.Wallet);
            mockWallet.id.returns('test-wallet-id');
            mockWallet.coin.returns(coin.getChain());
            mockWallet.coinSpecific.returns(undefined);
            // Mock getTransaction for RBF case
            if ('rbfTxIds' in resolvedTxParams) {
                const rbfTxParams = getTxParamsFromExplanation(explanation, { externalCustomChangeAddress: false });
                mockWallet.getTransaction.resolves({
                    outputs: rbfTxParams.recipients.map((r) => ({
                        valueString: typeof r.amount === 'string' ? r.amount : r.amount.toString(),
                        address: r.address,
                        // wallet field is undefined for external outputs (not self-sends)
                    })),
                });
            }
            // Mock verification options with keychains to disable networking
            // Use the same keychains that were used to create the PSBT
            const pubs = acidTest.rootWalletKeys.triple.map((k) => k.neutered().toBase58());
            const verification = {
                disableNetworking: true,
                keychains: {
                    user: { id: '0', pub: pubs[0], type: 'independent' },
                    backup: { id: '1', pub: pubs[1], type: 'independent' },
                    bitgo: { id: '2', pub: pubs[2], type: 'independent' },
                },
            };
            // Stub explainTransaction to return the explanation without making network calls
            stubExplainTransaction = sinon.stub(coin, 'explainTransaction').resolves(explanation);
            let txPrebuild;
            if (txFormat === 'psbt') {
                txPrebuild = {
                    txHex: psbt.toHex(),
                };
            }
            else if (txFormat === 'legacy') {
                txPrebuild = {
                    txHex: psbt.getUnsignedTx().toHex(),
                };
            }
            else {
                throw new Error(`Invalid txFormat: ${txFormat}`);
            }
            refParsedTransaction = await (0, parseTransaction_1.parseTransaction)(coin, {
                wallet: mockWallet,
                txParams: resolvedTxParams,
                txPrebuild,
                verification,
            });
        });
        after('cleanup', function () {
            if (stubExplainTransaction) {
                stubExplainTransaction.restore();
            }
        });
        it('should parse transaction without network calls', function () {
            strict_1.default.ok(refParsedTransaction);
            strict_1.default.ok(refParsedTransaction.keychains);
            strict_1.default.ok(refParsedTransaction.outputs);
        });
        it('should have valid keychains', function () {
            strict_1.default.ok(refParsedTransaction.keychains.user);
            strict_1.default.ok(refParsedTransaction.keychains.backup);
            strict_1.default.ok(refParsedTransaction.keychains.bitgo);
            const pubs = acidTest.rootWalletKeys.triple.map((k) => k.neutered().toBase58());
            strict_1.default.strictEqual(refParsedTransaction.keychains.user.pub, pubs[0]);
            strict_1.default.strictEqual(refParsedTransaction.keychains.backup.pub, pubs[1]);
            strict_1.default.strictEqual(refParsedTransaction.keychains.bitgo.pub, pubs[2]);
        });
        it('should have outputs classified as internal or external', function () {
            // Since we didn't specify any recipients, outputs will be classified based on whether they can be
            // verified as wallet addresses. Some may be external if address verification fails without a proper wallet setup.
            const totalOutputs = refParsedTransaction.outputs.length;
            const changeOutputs = refParsedTransaction.changeOutputs.length;
            const externalOutputs = refParsedTransaction.outputs.filter((o) => o.external === true).length;
            strict_1.default.strictEqual(externalOutputs, 3);
            strict_1.default.ok(totalOutputs > 0, 'should have at least one output');
            strict_1.default.strictEqual(changeOutputs + externalOutputs, totalOutputs, 'all outputs should be classified');
        });
        it('should have expected explicit and implicit external spend amounts', function () {
            strict_1.default.strictEqual(BigInt(refParsedTransaction.explicitExternalSpendAmount), expectedExplicitExternalSpendAmount);
            strict_1.default.strictEqual(BigInt(refParsedTransaction.implicitExternalSpendAmount), expectedImplicitExternalSpendAmount);
        });
    });
}
function describeTransaction(backend, filter = () => true) {
    describe(`parseTransaction (${backend})`, function () {
        utxolib.testutil.AcidTest.suite()
            .filter(filter)
            .forEach((test) => {
            // Default case: psbt format, infer recipients from explanation
            describeParseTransactionWith(test, 'default', backend, {
                txParams: 'inferFromExplanation',
                expectedExplicitExternalSpendAmount: 1800n,
                expectedImplicitExternalSpendAmount: 0n,
            });
            if (test.network !== utxolib.networks.bitcoin) {
                return;
            }
            // extended test suite for bitcoin
            describeParseTransactionWith(test, 'legacy', backend, {
                txFormat: 'legacy',
                txParams: 'inferFromExplanation',
                expectedExplicitExternalSpendAmount: 1800n,
                expectedImplicitExternalSpendAmount: 0n,
            });
            describeParseTransactionWith(test, 'empty recipients', backend, {
                txParams: {
                    recipients: [],
                },
                expectedExplicitExternalSpendAmount: 0n,
                expectedImplicitExternalSpendAmount: 1800n,
            });
            describeParseTransactionWith(test, 'rbf', backend, {
                txParams: {
                    rbfTxIds: ['PLACEHOLDER'],
                },
                expectedExplicitExternalSpendAmount: 1800n,
                expectedImplicitExternalSpendAmount: 0n,
            });
            describeParseTransactionWith(test, 'allowExternalChangeAddress', backend, {
                txParams: 'inferFromExplanation',
                externalCustomChangeAddress: true,
                expectedExplicitExternalSpendAmount: 1800n,
                expectedImplicitExternalSpendAmount: 0n,
            });
        });
    });
}
describeTransaction('utxolib');
describeTransaction('wasm');
