// @bitgo/abstract-utxo/dist/cjs/test/unit/transaction/fixedScript/signPsbt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const strict_1 = __importDefault(require("node:assert/strict"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const wasm_utxo_1 = require("@bitgo/wasm-utxo");
const decode_1 = require("../../../../src/transaction/decode");
const signPsbtUtxolib_1 = require("../../../../src/transaction/fixedScript/signPsbtUtxolib");
const signPsbtWasm_1 = require("../../../../src/transaction/fixedScript/signPsbtWasm");
const names_1 = require("../../../../src/names");
function getMockCoinUtxolib(keys) {
    return {
        async getMusig2Nonces(psbt, walletId) {
            psbt.setAllInputsMusig2NonceHD(keys.bitgo, { deterministic: true });
            return psbt;
        },
    };
}
function getMockCoinWasm(keys, network) {
    // Convert utxolib RootWalletKeys to wasm BIP32 using base58 string
    // This ensures the private key is properly transferred
    const bitgoXprv = keys.bitgo.toBase58();
    const bitgoKey = wasm_utxo_1.BIP32.fromBase58(bitgoXprv);
    const networkName = utxolib.getNetworkName(network);
    (0, strict_1.default)(networkName, 'network name is required');
    return {
        async getMusig2Nonces(psbt, walletId) {
            // Generate nonces using the bitgo key
            psbt.generateMusig2Nonces(bitgoKey);
            // Serialize and deserialize to simulate remote response
            // This creates a new object so we don't get "recursive use of an object" error
            return wasm_utxo_1.fixedScriptWallet.BitGoPsbt.fromBytes(psbt.serialize(), networkName);
        },
    };
}
function assertSignedUtxolib(psbt, userKey) {
    // Verify that all wallet inputs have been signed by user key
    psbt.data.inputs.forEach((input, inputIndex) => {
        const { scriptType } = utxolib.bitgo.parsePsbtInput(input);
        // Skip replay protection inputs (p2shP2pk)
        if (scriptType === 'p2shP2pk') {
            return;
        }
        // Verify user signature is present
        const isValid = psbt.validateSignaturesOfInputHD(inputIndex, userKey);
        (0, strict_1.default)(isValid, `input ${inputIndex} should have valid user signature`);
    });
}
function assertSignedWasm(psbt, userKey, rootWalletKeys, replayProtection) {
    const wasmUserKey = wasm_utxo_1.BIP32.from(userKey);
    const parsed = psbt.parseTransactionWithWalletKeys(rootWalletKeys, { replayProtection });
    // Verify that all wallet inputs have been signed by user key
    parsed.inputs.forEach((input, inputIndex) => {
        // Skip replay protection inputs (p2shP2pk)
        if (input.scriptType === 'p2shP2pk') {
            return;
        }
        // Verify user signature is present
        const isValid = psbt.verifySignature(inputIndex, wasmUserKey);
        (0, strict_1.default)(isValid, `input ${inputIndex} should have valid user signature (scriptType=${input.scriptType})`);
    });
}
function toWasmWalletKeys(keys) {
    return {
        triple: [keys.user, keys.backup, keys.bitgo],
        derivationPrefixes: keys.derivationPrefixes,
    };
}
function getReplayProtectionKeys(keys) {
    // Replay protection inputs use the underived user public key
    return {
        publicKeys: [keys.user.publicKey],
    };
}
function describeSignPsbtWithMusig2Participant(acidTest, { decodeWith }) {
    describe(`${acidTest.name} ${decodeWith}`, function () {
        it('should sign unsigned psbt to halfsigned', async function () {
            // Create unsigned PSBT
            const coinName = (0, names_1.getCoinName)(acidTest.network);
            const psbt = (0, decode_1.decodePsbtWith)(acidTest.createPsbt().toBuffer(), coinName, decodeWith);
            let result;
            if (decodeWith === 'utxolib') {
                (0, strict_1.default)(psbt instanceof utxolib.bitgo.UtxoPsbt, 'psbt should be a UtxoPsbt');
                result = await (0, signPsbtUtxolib_1.signPsbtWithMusig2ParticipantUtxolib)(getMockCoinUtxolib(acidTest.rootWalletKeys), psbt, acidTest.rootWalletKeys.user, {
                    signingStep: undefined,
                    walletId: 'test-wallet-id',
                });
                // Result should be a PSBT (not finalized)
                (0, strict_1.default)(result instanceof utxolib.bitgo.UtxoPsbt, 'should return UtxoPsbt when not last signature');
                assertSignedUtxolib(result, acidTest.rootWalletKeys.user);
            }
            else {
                (0, strict_1.default)(psbt instanceof wasm_utxo_1.fixedScriptWallet.BitGoPsbt, 'psbt should be a BitGoPsbt');
                const wasmWalletKeys = toWasmWalletKeys(acidTest.rootWalletKeys);
                const replayProtection = getReplayProtectionKeys(acidTest.rootWalletKeys);
                result = await (0, signPsbtWasm_1.signPsbtWithMusig2ParticipantWasm)(getMockCoinWasm(acidTest.rootWalletKeys, acidTest.network), psbt, acidTest.rootWalletKeys.user, wasm_utxo_1.fixedScriptWallet.RootWalletKeys.from(acidTest.rootWalletKeys), {
                    replayProtection,
                    signingStep: undefined,
                    walletId: 'test-wallet-id',
                });
                // Result should be a PSBT (not finalized)
                (0, strict_1.default)(result instanceof wasm_utxo_1.fixedScriptWallet.BitGoPsbt, 'should return BitGoPsbt when not last signature');
                assertSignedWasm(result, acidTest.rootWalletKeys.user, wasmWalletKeys, replayProtection);
            }
        });
    });
}
describe('signPsbtWithMusig2ParticipantUtxolib', function () {
    // Create test suite with includeP2trMusig2ScriptPath set to false
    // p2trMusig2 script path inputs are signed by user and backup keys,
    // which is not the typical signing pattern and makes testing more complex
    utxolib.testutil.AcidTest.suite({ includeP2trMusig2ScriptPath: false })
        .filter((test) => test.signStage === 'unsigned')
        .forEach((test) => {
        describeSignPsbtWithMusig2Participant(test, { decodeWith: 'utxolib' });
        describeSignPsbtWithMusig2Participant(test, { decodeWith: 'wasm-utxo' });
    });
});
