// @bitgo/utxo-core/dist/cjs/src/paygo/psbt/Errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMultiplePayGoProofAtPsbtIndex = exports.ErrorOutputIndexOutOfBounds = exports.ErrorPayGoAddressProofFailedVerification = exports.ErrorMultiplePayGoProof = exports.ErrorNoPayGoProof = exports.PayGoError = void 0;
class PayGoError extends Error {
    constructor(message) {
        super(message);
        this.name = this.constructor.name;
    }
}
exports.PayGoError = PayGoError;
class ErrorNoPayGoProof extends PayGoError {
    constructor(outputIndex) {
        super(`There is no paygo address proof encoded in the PSBT at output ${outputIndex}.`);
        this.outputIndex = outputIndex;
    }
}
exports.ErrorNoPayGoProof = ErrorNoPayGoProof;
class ErrorMultiplePayGoProof extends PayGoError {
    constructor() {
        super('There are multiple paygo address proofs encoded in the PSBT. Something went wrong.');
    }
}
exports.ErrorMultiplePayGoProof = ErrorMultiplePayGoProof;
class ErrorPayGoAddressProofFailedVerification extends PayGoError {
    constructor() {
        super('Cannot verify the paygo address signature with the provided pubkey.');
    }
}
exports.ErrorPayGoAddressProofFailedVerification = ErrorPayGoAddressProofFailedVerification;
class ErrorOutputIndexOutOfBounds extends PayGoError {
    constructor(outputIndex) {
        super(`Output index ${outputIndex} is out of bounds for PSBT outputs.`);
        this.outputIndex = outputIndex;
    }
}
exports.ErrorOutputIndexOutOfBounds = ErrorOutputIndexOutOfBounds;
class ErrorMultiplePayGoProofAtPsbtIndex extends PayGoError {
    constructor(outputIndex) {
        super(`There are multiple PayGo addresses in the PSBT output ${outputIndex}.`);
        this.outputIndex = outputIndex;
    }
}
exports.ErrorMultiplePayGoProofAtPsbtIndex = ErrorMultiplePayGoProofAtPsbtIndex;
