// @bitgo/sdk-coin-celo/dist/src/lib/resources.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mainnetCommon = exports.testnetCommon = void 0;
const statics_1 = require("@bitgo/statics");
const common_1 = __importDefault(require("@ethereumjs/common"));
/**
 * A Common object defining the chain and the hardfork for CELO Testnet
 */
exports.testnetCommon = common_1.default.custom({
    name: 'celo sepolia testnet',
    networkId: statics_1.coins.get('tcelo').network.chainId,
    chainId: statics_1.coins.get('tcelo').network.chainId,
}, {
    baseChain: 'sepolia',
    hardfork: 'london',
    eips: [1559],
});
/**
 * A Common object defining the chain and the hardfork for CELO Mainnet
 */
exports.mainnetCommon = common_1.default.custom({
    name: 'rc1',
    networkId: statics_1.coins.get('celo').network.chainId,
    chainId: statics_1.coins.get('celo').network.chainId,
}, { hardfork: 'london' });
