// @bitgo/sdk-coin-flrp/dist/test/resources/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ON_CHAIN_TEST_WALLET = exports.INVALID_LONG_KEYPAIR_PRV = exports.INVALID_PUBLIC_KEY_ERROR_MESSAGE = exports.INVALID_PRIVATE_KEY_ERROR_MESSAGE = exports.INVALID_SHORT_KEYPAIR_KEY = exports.CONTEXT = exports.ACCOUNT_4 = exports.ACCOUNT_3 = exports.ACCOUNT_2 = exports.ACCOUNT_1 = exports.SEED_ACCOUNT = void 0;
exports.SEED_ACCOUNT = {
    seed: '4c3b89f6ca897cb729d2146913877f71',
    privateKey: 'd8f07de5977843949bf4b81b0978117c85f2582c57d7dda9a0450ff3fbef19fe',
    publicKey: '03cd9197658fb563593560b93f8eba2aeb2e4f3781262569b3277b8da7f4f727b2',
    xPrivateKey: 'xprv9s21ZrQH143K3uPT3aSjQNaUYJQX4MyGXmavbDN5WBMAafS3PQ9XV2E5iMXLcNUppNPBh77UynnjMTL35t5BD8vuHqAYq8G3MNEbnEER3BY',
    xPublicKey: 'xpub661MyMwAqRbcGPTv9byjmWXD6LF1Tph7tzWXPbmh4Wt9TTmBvwTn2pYZZeQqnBEH6cTYCQEeYLLkvs7rwDN9cAKrK91rbBs8ixs532ZDZgE',
    flrpPrivateKey: 'PrivateKey-2eYRjENrQkjWdizt6PxxP1DPF3E2w6SYWaiSAMWJwDbqVWxMLW',
    addressMainnet: 'P-flare1uyp5n76gjqltrddur7qlrsmt3kyh8fnrrqwtal',
    addressTestnet: 'P-costwo1uyp5n76gjqltrddur7qlrsmt3kyh8fnrmwhqk7',
    message: 'test message',
    signature: '1692c0a25c84d389e63692ca3b7ebc89835c3b11f6e64a505ddd71664d2a3ae914e0d1be13a824ae97331805650a0631df9ae92ba133f9aa81911f3a56807ba301',
};
exports.ACCOUNT_1 = {
    seed: '4c3b89f6ca897cb729d2146913877f71',
    privateKey: 'a533d8419d4518e11cd8d9f049c73a8bdaf003d6602319f967ce3c243e646ba5',
    publicKey: '02a220e5fd108996d0e6c85db43384dcef8884bcaee1203e980f9f99f65ab3d3f3',
    xPrivateKey: 'xprv9s21ZrQH143K4SKxfadM7W3Yq1hhwJgGAFQio8sBqaXADVo8xmQiUe2cy6GwCkwsfDZMSe4W6Gv5vHTsaUF8yoYuzq8KFJPUu2p98RAXKsJ',
    xPublicKey: 'xpub661MyMwAqRbcGvQRmcAMUdzHP3YCLmQ7XULKbXGoPv496J8HWJiy2SM6pMHM89sQnyGiLF46dfzFB5ZRTwRmpiUq2hUkp5YccrcWG5XkS3D',
    addressMainnet: 'P-flare1evum5n0agffrdhg2dm2vg7svfgrra5ruxnmfwk',
    addressTestnet: 'P-costwo1evum5n0agffrdhg2dm2vg7svfgrra5ru7azz9h',
};
exports.ACCOUNT_2 = {
    seed: '4c3b89f6ca897cb729d2146913877f71',
    privateKey: '836d24396ff6e952e632b3552fc6591566e8be7c931a6c64fa7f685be9647841',
    publicKey: '03cd4660ac1570473e7107ed1f31425d81a624992bbc5983fd61787c560d8fd420',
    xPrivateKey: 'xprv9s21ZrQH143K2KjD8ytSfLWDfe2585pBJNdadLgwsEKoGLbGdHCKSK5yDnfcmToazd3oPLDXprtXnCvsn9T6MDJz1qwMPaq22oTrzqvyeDQ',
    xPublicKey: 'xprv9s21ZrQH143K2KjD8ytSfLWDfe2585pBJNdadLgwsEKoGLbGdHCKSK5yDnfcmToazd3oPLDXprtXnCvsn9T6MDJz1qwMPaq22oTrzqvyeDQ',
    addressMainnet: 'P-flare1xvngrrmqzldj50kpvqx7mhkx3htvh5x87nkl9j',
    addressTestnet: 'P-costwo1xvngrrmqzldj50kpvqx7mhkx3htvh5x8mhgsqy',
};
exports.ACCOUNT_3 = {
    seed: '4c3b89f6ca897cb729d2146913877f71',
    privateKey: '0906840926fcd038f42921709655e8b8b06613272a3ac9040510b7d4b26f09b6',
    publicKey: '029aa4d9f18d2994f2d833c3b9dd17a5ad9b1c3e744f0fe0b4cb1595f8f67fd12c',
    xPrivateKey: 'xprv9s21ZrQH143K4Tto9h1DW3gmMFpxGjSyUS3V1rCUjVEDDm8xzs38GCrsCryjEbrEovHhg12d55mBx4jHK1H19RyuhzJu4GQD9HBqLZCpA6b',
    xPublicKey: 'xpub661MyMwAqRbcGwyGFiYDsBdVuHfSgCApqey5pEc6HpmC6ZU7YQMNp1BM497dJjBnQZmNFPVQFRVLnG3yJvzyA8zGjseifmTq3HkQyq1zHqq',
    address: 'P-costwo1xqr8ps8s6qv5jke9wltc8dc9wm5q7ck2frt5hd',
};
exports.ACCOUNT_4 = {
    seed: '4c3b89f6ca897cb729d2146913877f71',
    privateKey: '6d19ef12622ad2e368806483f91445ba832a0cadd73d8585bfd9de59037e79b4',
    publicKey: '026bb5037dce5714dc9427e45eabebd689fc671b9c7ba7b0ffc74c944789ece9d1',
    xPrivateKey: 'xprv9s21ZrQH143K2rNMJAomSkL7i7rSudnEkrm4YWS4nAAPJAQKGqitmezqp5E38uVeLWkzLN7VXnkKy3uqnc3DDXqojQ2rXGLYBwtVRVcRe7d',
    xPublicKey: 'xpub661MyMwAqRbcFLSpQCLmotGrG9gwK6W685gfLtqgLVhNAxjTpP39KTKKfLFBxxptfe4k8KbyzWCPCFbm9xuSUXQNRQTNQoebYiupV8fSPHJ',
    address: 'P-costwo1jvjdvg6jdqez24c5kjxgsu47mqwvpyerk22yl8',
};
exports.CONTEXT = {
    xBlockchainID: 'FJuSwZuP85eyBpuBrKECnpPedGyXoDy2hP9q4JD8qBTZGxYbJ',
    pBlockchainID: '11111111111111111111111111111111LpoYY',
    cBlockchainID: 'vE8M98mEQH6wk56sStD1ML8HApTgSqfJZLk9gQ3Fsd4i6m3Bi',
    avaxAssetID: 'fxMAKpBQQpFedrUhWMsDYfCUJxdUw4mneTczKBzNg3rc2JUub',
    baseTxFee: 1000000n,
    createAssetTxFee: 1000000n,
    createSubnetTxFee: 100000000n,
    transformSubnetTxFee: 100000000n,
    createBlockchainTxFee: 100000000n,
    addPrimaryNetworkValidatorFee: 0n,
    addPrimaryNetworkDelegatorFee: 0n,
    addSubnetValidatorFee: 1000000n,
    addSubnetDelegatorFee: 1000000n,
    networkID: 114,
    hrp: 'costwo',
    platformFeeConfig: {
        weights: { 0: 1, 1: 1000, 2: 1000, 3: 4 },
        maxCapacity: 1000000n,
        maxPerSecond: 100000n,
        targetPerSecond: 50000n,
        minPrice: 250n,
        excessConversionConstant: 2164043n,
    },
};
exports.INVALID_SHORT_KEYPAIR_KEY = '82A34E';
exports.INVALID_PRIVATE_KEY_ERROR_MESSAGE = 'Unsupported private key';
exports.INVALID_PUBLIC_KEY_ERROR_MESSAGE = 'Unsupported public key';
exports.INVALID_LONG_KEYPAIR_PRV = exports.SEED_ACCOUNT.privateKey + 'F1';
/**
 * On-chain verified test wallet keys
 * Used for testing transactions that have been validated on Flare Coston2 testnet
 */
exports.ON_CHAIN_TEST_WALLET = {
    user: {
        privateKey: 'c9e1638b866930153d06d30234ab80b34cd8af52715c881cb5a3b60a7aad3fef',
        pChainAddress: 'P-costwo1cgpnd2v0a0qxevncn4m7st2lychuyn7czpd5k9',
        corethAddress: 'C-costwo1cgpnd2v0a0qxevncn4m7st2lychuyn7czpd5k9',
    },
    bitgo: {
        privateKey: '854309c49c8064a806b1d95964b35d23a3b01b4f8dd16884348f37669af6a0ed',
        pChainAddress: 'P-costwo1xz2dh8g69l7almnks86l47m7dand9pzfnfpxwp',
        corethAddress: 'C-costwo1xz2dh8g69l7almnks86l47m7dand9pzfnfpxwp',
    },
    backup: {
        privateKey: '00bd5a0bfdb5e4b536c4ccdf181014bfb8b022d49e563f48fc799b9d2f20def0',
        pChainAddress: 'P-costwo107q8ch4jsa58xdjlhl7ggazu6vgqgekhuszr7a',
        corethAddress: 'C-costwo107q8ch4jsa58xdjlhl7ggazu6vgqgekhuszr7a',
    },
};
