// @bitgo/sdk-coin-ada/dist/src/adaToken.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdaToken = void 0;
const ada_1 = require("./ada");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const lib_1 = require("./lib");
const CardanoWasm = __importStar(require("@emurgo/cardano-serialization-lib-nodejs"));
const assert_1 = __importDefault(require("assert"));
class AdaToken extends ada_1.Ada {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('ada') : statics_1.coins.get('tada');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new AdaToken(bitgo, config);
    }
    static createTokenConstructors(tokenConfigs = [...statics_1.tokens.bitcoin.ada.tokens, ...statics_1.tokens.testnet.ada.tokens]) {
        const tokensCtors = [];
        for (const token of tokenConfigs) {
            const tokenConstructor = AdaToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get policyId() {
        return this.tokenConfig.policyId;
    }
    get assetName() {
        return this.tokenConfig.assetName;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Cardano Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return false;
    }
    /**
     * Flag for sending data along with transactions
     * @returns {boolean} True if okay to send tx data (CELO), false otherwise
     */
    transactionDataAllowed() {
        return false;
    }
    get contractAddress() {
        return this.tokenConfig.contractAddress;
    }
    /**
     * Verify that a token transaction prebuild complies with the original intention.
     * For token transfers, we need to verify the token amount in multiAssets, not the ADA amount.
     * For token consolidation, we verify all outputs go to the base address.
     *
     * @param params.txPrebuild prebuild transaction
     * @param params.txParams transaction parameters
     * @param params.verification verification options (includes consolidationToBaseAddress flag)
     * @param params.wallet wallet object for getting base address
     * @return true if verification success
     */
    async verifyTransaction(params) {
        try {
            const coinConfig = statics_1.coins.get(this.getBaseChain());
            const { txPrebuild, txParams, verification, wallet } = params;
            const transaction = new lib_1.Transaction(coinConfig);
            (0, assert_1.default)(txPrebuild.txHex, new Error('missing required tx prebuild property txHex'));
            transaction.fromRawTransaction(txPrebuild.txHex);
            const txJson = transaction.toJson();
            if (txParams.recipients !== undefined) {
                // assetName in tokenConfig is ASCII (e.g. 'WATER'), convert to hex for comparison
                const asciiEncodedAssetName = Buffer.from(this.tokenConfig.assetName).toString('hex');
                // ASCII encoded asset name may be appended to the policy ID (consistent with crypto compare)
                // But cardano sdk requires only the policy ID (28 bytes = 56 hex chars) for ScriptHash
                let policyId = this.tokenConfig.policyId;
                if (policyId.endsWith(asciiEncodedAssetName)) {
                    policyId = policyId.substring(0, policyId.length - asciiEncodedAssetName.length);
                }
                const policyScriptHash = CardanoWasm.ScriptHash.from_hex(policyId);
                const assetName = CardanoWasm.AssetName.new(Buffer.from(asciiEncodedAssetName, 'hex'));
                for (const recipient of txParams.recipients) {
                    const found = txJson.outputs.some((output) => {
                        if (recipient.address !== output.address || !output.multiAssets) {
                            return false;
                        }
                        const multiAssets = output.multiAssets;
                        const tokenQty = multiAssets.get_asset(policyScriptHash, assetName);
                        return tokenQty && tokenQty.to_str() === recipient.amount;
                    });
                    if (!found) {
                        throw new Error('cannot find recipient in expected output');
                    }
                }
            }
            else if (verification?.consolidationToBaseAddress) {
                // For token consolidation, verify all outputs go to the base address
                const baseAddress = wallet?.coinSpecific()?.baseAddress || wallet?.coinSpecific()?.rootAddress;
                for (const output of txJson.outputs) {
                    if (output.address !== baseAddress) {
                        throw new Error('tx outputs does not match with expected address');
                    }
                }
            }
        }
        catch (e) {
            if (e instanceof sdk_core_1.NodeEnvironmentError) {
                return true;
            }
            throw e;
        }
        return true;
    }
}
exports.AdaToken = AdaToken;
