// @bitgo/sdk-coin-polyx/dist/test/unit/transactionBuilder/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildMainnetConfig = exports.buildTestConfig = void 0;
const sinon_1 = __importDefault(require("sinon"));
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const statics_1 = require("@bitgo/statics");
const sdk_core_1 = require("@bitgo/sdk-core");
const lib_1 = require("../../../src/lib");
const src_1 = require("../../../src/");
const resources_1 = require("../../resources");
const resources_2 = require("../../../src/resources");
const buildTestConfig = () => {
    return statics_1.coins.get('tpolyx');
};
exports.buildTestConfig = buildTestConfig;
const buildMainnetConfig = () => {
    return statics_1.coins.get('polyx');
};
exports.buildMainnetConfig = buildMainnetConfig;
class StubTransactionBuilder extends lib_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.material(src_1.utils.getMaterial(_coinConfig.network.type));
    }
    validateDecodedTransaction(decodedTxn) {
        return;
    }
    buildTransaction() {
        throw new Error('Method not implemented.');
    }
    getSender() {
        return this._sender;
    }
    getReferenceBlock() {
        return this._referenceBlock;
    }
    getNonce() {
        return this._nonce;
    }
    getTip() {
        return this._tip;
    }
    getEraPeriod() {
        return this._eraPeriod;
    }
    buildImplementation() {
        return super.buildImplementation();
    }
    fromImplementation(rawTransaction) {
        return super.fromImplementation(rawTransaction);
    }
    signImplementation(key) {
        return super.signImplementation(key);
    }
    get transactionType() {
        throw new Error('Method not implemented.');
    }
    getTransaction() {
        return this._transaction;
    }
    getMaterial() {
        return this._material;
    }
}
describe('Polyx Transfer Builder Base', () => {
    let builder;
    const sender = resources_1.accounts.account1;
    const receiver = resources_1.accounts.account2;
    beforeEach(() => {
        builder = new StubTransactionBuilder((0, exports.buildTestConfig)());
    });
    describe('setter validation', () => {
        it('should validate sender address', () => {
            const spy = sinon_1.default.spy(builder, 'validateAddress');
            assert_1.default.throws(() => builder.sender({ address: 'asd' }), (e) => e.message === `The address 'asd' is not a well-formed dot address`);
            should_1.default.doesNotThrow(() => builder.sender({ address: sender.address }));
            sinon_1.default.assert.calledTwice(spy);
        });
        it('should validate eraPeriod', () => {
            const spy = sinon_1.default.spy(builder, 'validateValue');
            assert_1.default.throws(() => builder.validity({ maxDuration: -1 }), (e) => e.message === 'Value cannot be less than zero');
            should_1.default.doesNotThrow(() => builder.validity({ maxDuration: 64 }));
            sinon_1.default.assert.calledTwice(spy);
        });
        it('should validate nonce', () => {
            const spy = sinon_1.default.spy(builder, 'validateValue');
            assert_1.default.throws(() => builder.sequenceId({ name: 'Nonce', keyword: 'nonce', value: -1 }), (e) => e.message === 'Value cannot be less than zero');
            should_1.default.doesNotThrow(() => builder.sequenceId({ name: 'Nonce', keyword: 'nonce', value: 10 }));
            sinon_1.default.assert.calledTwice(spy);
        });
        it('should validate tip', () => {
            const spy = sinon_1.default.spy(builder, 'validateValue');
            assert_1.default.throws(() => builder.fee({ amount: -1, type: 'tip' }), (e) => e.message === 'Value cannot be less than zero');
            should_1.default.doesNotThrow(() => builder.fee({ amount: 10, type: 'tip' }));
            sinon_1.default.assert.calledTwice(spy);
        });
        it('should validate blockNumber', () => {
            const spy = sinon_1.default.spy(builder, 'validateValue');
            assert_1.default.throws(() => builder.validity({ firstValid: -1 }), (e) => e.message === 'Value cannot be less than zero');
            should_1.default.doesNotThrow(() => builder.validity({ firstValid: 10 }));
            sinon_1.default.assert.calledTwice(spy);
        });
    });
    describe('build base transaction', () => {
        it('should build validate base fields', async () => {
            builder
                .material(src_1.utils.getMaterial((0, exports.buildTestConfig)().network.type))
                .sender({ address: sender.address })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            should_1.default.doesNotThrow(() => builder.validateTransaction(builder.getTransaction()));
        });
        it('should build a base transaction on testnet', async () => {
            const material = builder.getMaterial();
            should_1.default.deepEqual(material.specName, resources_2.testnetMaterial.specName);
            should_1.default.deepEqual(material.genesisHash, resources_2.testnetMaterial.genesisHash);
            should_1.default.deepEqual(material.specVersion, resources_2.testnetMaterial.specVersion);
            should_1.default.deepEqual(material.chainName, resources_2.testnetMaterial.chainName);
        });
        it('should build from raw signed tx', async () => {
            builder.from(resources_1.rawTx.transfer.signed);
            should_1.default.deepEqual(builder.getSender(), '5CLYvxwx4PUS678MNuhNJ9EfpUU9utrYCz9WVxovac4u9AYD');
            should_1.default.deepEqual(builder.getNonce(), 34);
            should_1.default.deepEqual(builder.getEraPeriod(), 64);
            should_1.default.deepEqual(builder.getTip(), undefined);
        });
        it('should build from raw unsigned tx', async () => {
            builder.from(resources_1.rawTx.transfer.unsigned);
            should_1.default.deepEqual(builder.getReferenceBlock(), '0xe5c5396f3291d2575d138b98d34eac2fc156e5280276f57a30a8e3b8fbc49bfc');
            should_1.default.deepEqual(builder.getNonce(), 36);
            should_1.default.deepEqual(builder.getEraPeriod(), 64);
            should_1.default.deepEqual(builder.getTip(), undefined);
        });
    });
    describe('add TSS signature', function () {
        let MPC;
        before('initialize mpc module', async () => {
            MPC = await sdk_core_1.Eddsa.initialize();
        });
        it('should add TSS signature', async () => {
            const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get('tpolyx'));
            const A = MPC.keyShare(1, 2, 3);
            const B = MPC.keyShare(2, 2, 3);
            const C = MPC.keyShare(3, 2, 3);
            const A_combine = MPC.keyCombine(A.uShare, [B.yShares[1], C.yShares[1]]);
            const B_combine = MPC.keyCombine(B.uShare, [A.yShares[2], C.yShares[2]]);
            const C_combine = MPC.keyCombine(C.uShare, [A.yShares[3], B.yShares[3]]);
            const commonPub = A_combine.pShare.y;
            const dotKeyPair = new lib_1.KeyPair({ pub: commonPub });
            const sender = dotKeyPair.getAddress(sdk_core_1.DotAddressFormat.substrate);
            let transferBuilder = factory
                .getTransferBuilder()
                .amount('90034235235322')
                .to({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .sender({ address: sender })
                .to({ address: receiver.address })
                .memo('0')
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            const unsignedTransaction = await transferBuilder.build();
            const signablePayload = unsignedTransaction.signablePayload;
            // signing with A and B
            let A_sign_share = MPC.signShare(signablePayload, A_combine.pShare, [A_combine.jShares[2]]);
            let B_sign_share = MPC.signShare(signablePayload, B_combine.pShare, [B_combine.jShares[1]]);
            let A_sign = MPC.sign(signablePayload, A_sign_share.xShare, [B_sign_share.rShares[1]], [C.yShares[1]]);
            let B_sign = MPC.sign(signablePayload, B_sign_share.xShare, [A_sign_share.rShares[2]], [C.yShares[2]]);
            // sign the message_buffer (unsigned txHex)
            let signature = MPC.signCombine([A_sign, B_sign]);
            let rawSignature = Buffer.concat([Buffer.from(signature.R, 'hex'), Buffer.from(signature.sigma, 'hex')]);
            transferBuilder = factory
                .getTransferBuilder()
                .amount('90034235235322')
                .to({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .sender({ address: sender })
                .to({ address: receiver.address })
                .memo('0')
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            transferBuilder.addSignature({ pub: dotKeyPair.getKeys().pub }, rawSignature);
            let signedTransaction = await transferBuilder.build();
            signedTransaction.signature.length.should.equal(1);
            signedTransaction.signature[0].should.equal(rawSignature.toString('hex'));
            // signing with A and C
            A_sign_share = MPC.signShare(signablePayload, A_combine.pShare, [A_combine.jShares[3]]);
            let C_sign_share = MPC.signShare(signablePayload, C_combine.pShare, [C_combine.jShares[1]]);
            A_sign = MPC.sign(signablePayload, A_sign_share.xShare, [C_sign_share.rShares[1]], [B.yShares[1]]);
            let C_sign = MPC.sign(signablePayload, C_sign_share.xShare, [A_sign_share.rShares[3]], [B.yShares[3]]);
            signature = MPC.signCombine([A_sign, C_sign]);
            rawSignature = Buffer.concat([Buffer.from(signature.R, 'hex'), Buffer.from(signature.sigma, 'hex')]);
            transferBuilder = factory
                .getTransferBuilder()
                .amount('90034235235322')
                .to({ address: '5Ffp1wJCPu4hzVDTo7XaMLqZSvSadyUQmxWPDw74CBjECSoq' })
                .sender({ address: sender })
                .to({ address: receiver.address })
                .memo('0')
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            transferBuilder.addSignature({ pub: dotKeyPair.getKeys().pub }, rawSignature);
            signedTransaction = await transferBuilder.build();
            signedTransaction.signature.length.should.equal(1);
            signedTransaction.signature[0].should.equal(rawSignature.toString('hex'));
            // signing with B and C
            B_sign_share = MPC.signShare(signablePayload, B_combine.pShare, [B_combine.jShares[3]]);
            C_sign_share = MPC.signShare(signablePayload, C_combine.pShare, [C_combine.jShares[2]]);
            B_sign = MPC.sign(signablePayload, B_sign_share.xShare, [C_sign_share.rShares[2]], [A.yShares[2]]);
            C_sign = MPC.sign(signablePayload, C_sign_share.xShare, [B_sign_share.rShares[3]], [A.yShares[3]]);
            signature = MPC.signCombine([B_sign, C_sign]);
            rawSignature = Buffer.concat([Buffer.from(signature.R, 'hex'), Buffer.from(signature.sigma, 'hex')]);
            transferBuilder = factory
                .getTransferBuilder()
                .amount('90034235235322')
                .sender({ address: sender })
                .to({ address: receiver.address })
                .memo('0')
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 })
                .fee({ amount: 0, type: 'tip' });
            transferBuilder.addSignature({ pub: dotKeyPair.getKeys().pub }, rawSignature);
            signedTransaction = await transferBuilder.build();
            signedTransaction.signature.length.should.equal(1);
            signedTransaction.signature[0].should.equal(rawSignature.toString('hex'));
            const rebuiltTransaction = await factory
                .from(signedTransaction.toBroadcastFormat())
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .build();
            rebuiltTransaction.signature[0].should.equal(rawSignature.toString('hex'));
        });
    });
});
