// @bitgo/sdk-coin-polyx/dist/test/unit/transactionBuilder/batchStakingBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const statics_1 = require("@bitgo/statics");
const should_1 = __importDefault(require("should"));
const sinon_1 = __importDefault(require("sinon"));
const lib_1 = require("../../../src/lib");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const resources_1 = require("../../resources");
function createMockTransaction(txData) {
    return {
        id: '123',
        type: sdk_core_1.TransactionType.Batch,
        toBroadcastFormat: () => txData,
        inputs: [],
        outputs: [],
        signature: ['mock-signature'],
        toJson: () => ({
            id: '123',
            type: 'Batch',
            sender: resources_1.accounts.account1.address,
            referenceBlock: '0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d',
            blockNumber: 100,
            genesisHash: '0x',
            nonce: 1,
            tip: 0,
            specVersion: 1,
            transactionVersion: 1,
            chainName: 'Polymesh',
            inputs: [],
            outputs: [],
        }),
    };
}
describe('Polyx Batch Builder', function () {
    let builder;
    const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get('tpolyx'));
    const senderAddress = resources_1.accounts.account1.address;
    const controllerAddress = resources_1.accounts.account2.address;
    const validatorAddress = '5C7kNpSvVr22Z1X6gVAUjfahSJfSpvw4DHNoY7uUHpLfEJZR';
    const testAmount = '10000';
    beforeEach(() => {
        builder = factory.getBatchBuilder();
    });
    describe('setter validation', () => {
        it('should validate staking amount', () => {
            const spy = sinon_1.default.spy(builder, 'validateValue');
            should_1.default.throws(() => builder.amount('-1'), /Value cannot be less than zero/);
            should_1.default.doesNotThrow(() => builder.amount('1000'));
            sinon_1.default.assert.calledTwice(spy);
        });
        it('should validate controller address', () => {
            const spy = sinon_1.default.spy(builder, 'validateAddress');
            should_1.default.throws(() => builder.controller({ address: 'invalid-address' }), /is not a well-formed/);
            should_1.default.doesNotThrow(() => builder.controller({ address: controllerAddress }));
            sinon_1.default.assert.calledTwice(spy);
        });
        it('should validate validator addresses', () => {
            const spy = sinon_1.default.spy(builder, 'validateAddress');
            should_1.default.throws(() => builder.validators(['invalid-address']), /is not a well-formed/);
            should_1.default.doesNotThrow(() => builder.validators([validatorAddress]));
            sinon_1.default.assert.calledTwice(spy);
        });
    });
    describe('Bond and Nominate', function () {
        it('should validate bond and nominate args together', () => {
            builder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .validators([validatorAddress]);
            should_1.default.doesNotThrow(() => {
                builder.testValidateFields();
            });
            builder = factory.getBatchBuilder();
            builder.amount(testAmount).controller({ address: controllerAddress }).payee('Staked');
            should_1.default.throws(() => {
                builder.testValidateFields();
            }, /must include both bond and nominate operations/);
            builder = factory.getBatchBuilder();
            builder.validators([validatorAddress]);
            should_1.default.throws(() => {
                builder.testValidateFields();
            }, /must include both bond and nominate operations/);
        });
        it('should build a batch transaction with both bond and nominate', async () => {
            builder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .validators([validatorAddress])
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            const mockTx = createMockTransaction(resources_1.stakingTx.batch.bondAndNominate.unsigned);
            sinon_1.default.stub(builder, 'build').resolves(mockTx);
            const tx = await builder.build();
            should_1.default.exist(tx);
            should_1.default.equal(builder.getAmount(), testAmount);
            should_1.default.equal(builder.getController(), controllerAddress);
            should_1.default.equal(builder.getPayee(), 'Staked');
            should_1.default.deepEqual(builder.getValidators(), [validatorAddress]);
        });
    });
    describe('Transaction Validation', function () {
        it('should build, decode, and validate a real batch staking transaction', () => {
            // Build the transaction with real parameters
            builder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .validators([validatorAddress])
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            // Set up material for decoding
            const material = utils_1.default.getMaterial(statics_1.coins.get('tpolyx').network.type);
            builder.material(material);
            // Build the actual unsigned transaction
            const unsignedTx = builder['buildTransaction']();
            const registry = builder['_registry'];
            // Decode the actual built transaction
            const decodedTx = (0, txwrapper_polkadot_1.decode)(unsignedTx, {
                metadataRpc: material.metadata,
                registry: registry,
            });
            // Validate the decoded transaction structure
            should_1.default.equal(decodedTx.method.name, 'batchAll');
            should_1.default.equal(decodedTx.method.pallet, 'utility');
            const batchArgs = decodedTx.method.args;
            should_1.default.exist(batchArgs.calls);
            should_1.default.equal(batchArgs.calls.length, 2);
            // Validate first call is bond
            const firstCall = batchArgs.calls[0];
            const firstCallMethod = utils_1.default.decodeMethodName(firstCall, registry);
            should_1.default.equal(firstCallMethod, 'bond');
            const bondArgs = firstCall.args;
            should_1.default.equal(bondArgs.value, testAmount);
            // Controller can be either a string or an object with id property
            const controllerValue = typeof bondArgs.controller === 'string' ? bondArgs.controller : bondArgs.controller.id;
            should_1.default.equal(controllerValue, controllerAddress);
            // Payee can be either a string or an object with staked property
            const payeeValue = typeof bondArgs.payee === 'string'
                ? bondArgs.payee
                : bondArgs.payee.staked !== undefined
                    ? 'Staked'
                    : bondArgs.payee;
            should_1.default.equal(payeeValue, 'Staked');
            // Validate second call is nominate
            const secondCall = batchArgs.calls[1];
            const secondCallMethod = utils_1.default.decodeMethodName(secondCall, registry);
            should_1.default.equal(secondCallMethod, 'nominate');
            const nominateArgs = secondCall.args;
            // Targets can be either strings or objects with id property
            const targetValues = nominateArgs.targets.map((target) => typeof target === 'string' ? target : target.id);
            should_1.default.deepEqual(targetValues, [validatorAddress]);
            // Now validate using the builder's validation method
            should_1.default.doesNotThrow(() => {
                builder.validateDecodedTransaction(decodedTx);
            });
        });
        it('should reject invalid batch transaction when built', () => {
            // Try to build a batch with only bond (missing nominate)
            builder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            // Set up material
            const material = utils_1.default.getMaterial(statics_1.coins.get('tpolyx').network.type);
            builder.material(material);
            // Should throw when trying to build without validators
            should_1.default.throws(() => {
                builder['buildTransaction']();
            }, /must include both bond and nominate operations/);
        });
        it('should reject validation of non-batch method name', () => {
            // Create a mock decoded transaction with wrong method name
            const mockDecodedTx = {
                method: {
                    name: 'bond',
                    pallet: 'staking',
                    args: {},
                },
            };
            should_1.default.throws(() => {
                builder.validateDecodedTransaction(mockDecodedTx);
            }, /Invalid transaction type/);
        });
    });
    describe('From Raw Transaction', function () {
        it('should rebuild from real batch transaction', async () => {
            // First build a transaction to get a real raw transaction
            const originalBuilder = factory.getBatchBuilder();
            originalBuilder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .validators([validatorAddress])
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            // Set up material
            const material = utils_1.default.getMaterial(statics_1.coins.get('tpolyx').network.type);
            originalBuilder.material(material);
            // Build the transaction and get the serialized hex
            const tx = await originalBuilder.build();
            const rawTxHex = tx.toBroadcastFormat();
            // Create a new builder and reconstruct from the transaction hex
            const newBuilder = factory.getBatchBuilder();
            newBuilder.material(material);
            newBuilder.from(rawTxHex);
            // Verify the reconstructed builder has the same parameters
            should_1.default.equal(newBuilder.getAmount(), testAmount);
            should_1.default.equal(newBuilder.getController(), controllerAddress);
            should_1.default.equal(newBuilder.getPayee(), 'Staked');
            should_1.default.deepEqual(newBuilder.getValidators(), [validatorAddress]);
        });
    });
    describe('should create properly configured transactions', function () {
        it('a batch transaction', async function () {
            builder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .validators([validatorAddress])
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            const mockTx = createMockTransaction(resources_1.stakingTx.batch.bondAndNominate.unsigned);
            sinon_1.default.stub(builder, 'build').resolves(mockTx);
            const tx = await builder.build();
            should_1.default.exist(tx);
        });
        it('a batch transaction with custom replay protection', async function () {
            builder
                .amount(testAmount)
                .controller({ address: controllerAddress })
                .payee('Staked')
                .validators([validatorAddress])
                .sender({ address: senderAddress })
                .validity({ firstValid: 4000, maxDuration: 100 })
                .referenceBlock('0x249799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754e')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 200 });
            const mockTx = createMockTransaction('custom-tx-format');
            sinon_1.default.stub(builder, 'build').resolves(mockTx);
            const tx = await builder.build();
            should_1.default.exist(tx);
        });
    });
});
