// @bitgo/sdk-coin-polyx/dist/test/unit/transactionBuilder/batchUnstakingBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const statics_1 = require("@bitgo/statics");
const should_1 = __importDefault(require("should"));
const sinon_1 = __importDefault(require("sinon"));
const lib_1 = require("../../../src/lib");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../../../src/lib/utils"));
const resources_1 = require("../../resources");
function createMockTransaction(txData) {
    return {
        id: '123',
        type: sdk_core_1.TransactionType.Batch,
        toBroadcastFormat: () => txData,
        inputs: [],
        outputs: [],
        signature: ['mock-signature'],
        toJson: () => ({
            id: '123',
            type: 'Batch',
            sender: resources_1.accounts.account1.address,
            referenceBlock: '0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d',
            blockNumber: 100,
            genesisHash: '0x',
            nonce: 1,
            tip: 0,
            specVersion: 1,
            transactionVersion: 1,
            chainName: 'Polymesh',
            inputs: [],
            outputs: [],
        }),
    };
}
describe('Polyx BatchUnstaking Builder', function () {
    let builder;
    const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get('tpolyx'));
    const senderAddress = resources_1.accounts.account1.address;
    const testAmount = '10000';
    beforeEach(() => {
        builder = factory.getBatchUnstakingBuilder();
    });
    describe('setter validation', () => {
        it('should validate unstaking amount', () => {
            const spy = sinon_1.default.spy(builder, 'validateValue');
            should_1.default.throws(() => builder.amount('-1'), /Value cannot be less than zero/);
            should_1.default.doesNotThrow(() => builder.amount('1000'));
            sinon_1.default.assert.calledTwice(spy);
        });
    });
    describe('Build and Sign', function () {
        it('should build a batch unstaking transaction', async () => {
            builder
                .amount(testAmount)
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            const mockTx = createMockTransaction(resources_1.rawTx.unstake.unsigned);
            sinon_1.default.stub(builder, 'build').resolves(mockTx);
            const tx = await builder.build();
            should_1.default.exist(tx);
            should_1.default.equal(builder.getAmount(), testAmount);
        });
    });
    describe('Transaction Validation', function () {
        it('should build, decode, and validate a real batch unstaking transaction', () => {
            // Build the transaction with real parameters
            builder
                .amount(testAmount)
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            // Set up material for decoding
            const material = utils_1.default.getMaterial(statics_1.coins.get('tpolyx').network.type);
            builder.material(material);
            // Build the actual unsigned transaction
            const unsignedTx = builder['buildTransaction']();
            const registry = builder['_registry'];
            // Decode the actual built transaction
            const decodedTx = (0, txwrapper_polkadot_1.decode)(unsignedTx, {
                metadataRpc: material.metadata,
                registry: registry,
            });
            // Validate the decoded transaction structure
            should_1.default.equal(decodedTx.method.name, 'batchAll');
            should_1.default.equal(decodedTx.method.pallet, 'utility');
            const batchArgs = decodedTx.method.args;
            should_1.default.exist(batchArgs.calls);
            should_1.default.equal(batchArgs.calls.length, 2);
            // Validate first call is chill
            const firstCall = batchArgs.calls[0];
            const firstCallMethod = utils_1.default.decodeMethodName(firstCall, registry);
            should_1.default.equal(firstCallMethod, 'chill');
            // Validate second call is unbond
            const secondCall = batchArgs.calls[1];
            const secondCallMethod = utils_1.default.decodeMethodName(secondCall, registry);
            should_1.default.equal(secondCallMethod, 'unbond');
            const unbondArgs = secondCall.args;
            should_1.default.equal(unbondArgs.value, testAmount);
            // Now validate using the builder's validation method
            should_1.default.doesNotThrow(() => {
                builder.validateDecodedTransaction(decodedTx);
            });
        });
        it('should reject invalid transaction types', () => {
            const mockDecodedTx = {
                method: {
                    name: 'transfer',
                    pallet: 'balances',
                    args: {},
                },
            };
            should_1.default.throws(() => {
                builder.validateDecodedTransaction(mockDecodedTx);
            }, /Invalid transaction type/);
        });
        it('should validate amount is positive', () => {
            builder.amount(testAmount);
            should_1.default.doesNotThrow(() => {
                builder.testValidateFields();
            });
            should_1.default.throws(() => {
                builder.amount('-10');
                builder.testValidateFields();
            }, /Value cannot be less than zero/);
        });
    });
    describe('From Raw Transaction', function () {
        it('should rebuild from real batch unstaking transaction', async () => {
            // First build a transaction to get a real raw transaction
            const originalBuilder = factory.getBatchUnstakingBuilder();
            originalBuilder
                .amount(testAmount)
                .sender({ address: senderAddress })
                .validity({ firstValid: 3933, maxDuration: 64 })
                .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
                .sequenceId({ name: 'Nonce', keyword: 'nonce', value: 100 });
            // Set up material
            const material = utils_1.default.getMaterial(statics_1.coins.get('tpolyx').network.type);
            originalBuilder.material(material);
            // Build the transaction and get the serialized hex
            const tx = await originalBuilder.build();
            const rawTxHex = tx.toBroadcastFormat();
            // Create a new builder and reconstruct from the transaction hex
            const newBuilder = factory.getBatchUnstakingBuilder();
            newBuilder.material(material);
            newBuilder.from(rawTxHex);
            // Verify the reconstructed builder has the same parameters
            should_1.default.equal(newBuilder.getAmount(), testAmount);
        });
    });
});
