// @bitgo/sdk-coin-polyx/dist/test/resources/wrwUsers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testnetBlock = exports.unsignedSweepUser = exports.wrwUser = void 0;
exports.wrwUser = {
    userKey: '{"iv":"TR25BHPM0F8OrlhC5UEnUA==","v":1,"iter":10000,"ks":256,"ts":64,"mode"\n' +
        ':"ccm","adata":"","cipher":"aes","salt":"uQfN/R/hYaM=","ct":"IzM4WznffhakZk\n' +
        'KwsRxTEcQ7PIcr5f6ro2YXbn1mP/dlMaYlmGOE1fJemchVeuFUWLGphJPMaUz3v4lJbklvRGtTP\n' +
        'F8+JeFSp/ps++wirzlZoAmPOGrOPRg7TkOJXR7xZyBJHeUoV9xu7X1iKGrxhYr70CrzlkrYqlQR\n' +
        'qU8NiV1YbeY+Ra56m9RfeE2HaXk2s9xNzn8mfdOdjkd1sjXwTzv3WPTBWPD35H8EiDbLuVLLOBg\n' +
        'tVPNahBimxffWE5B6keQTN3ezKNTwE0z3B4sBTGHb8wbumhRAO42/BV7MNbvCDPe+GfkqljMu13\n' +
        'u9PAaRb3Y/nK1szIxqN4PpUYP92UYr/QjveuHk37+wxQhljfn81jhkhtGOvnv1EaBhZNJ0PWbrf\n' +
        'E9Xprx8vLmyy8UiGayUvXka2bEZyPksjBpD7chhKbqCj3q5EhJQLUiydLr9/BV50seM1+AQ9w+R\n' +
        '8KjybUgwIF/qTLcaCxVHeAKRzs6LMGj43+l1kTkdaGRre+6MAK2aExZK2FJer49v8htKo7Z6vy+\n' +
        '149DBwTnvCGEDQVXuumKCGX0CCfzbLAMZVKB/Xqdd3e9B99QbVFmPQpeAlJXrOpaoXS90xSSBKs\n' +
        'dbr1wkRHlYB/NvQZwl2sGfIcPJASfWKLFUSXJM7ql1y89llNNSW7RUBcf9tKK5SqmVyGxWZ+1LJ\n' +
        'a5ZEh4BRYh5Kx4IleCQNyZI73XVZE8kyHo07TWwlI8jR8DUIsuF1cBmFRKn2XoikQrDYie/o1zS\n' +
        '1hDMzvF4C9BeacW6iPZ1n9/0OkwgiUIDJGh4vIOei/9U/N2scc8z5+F/yaXXW7COKxqKkx74tqI\n' +
        'ohPm0CTnH2t2cqjK+C5x3FdwiPT+uIzgSU04EZP4gzngZ11yunj7cvw80x9/BK+GwbM97haJ8Je\n' +
        'lMDg5++m6fJlb6weW1Rd6IiUfRSXpio7KQbxZ04UG5uPROAY1xbCJbGQi9OjcyXQKQtL81WkHSr\n' +
        'LGjN/jXCts7OH3xZRaZk66DP6X3LaT0b6qmlem2Uvf+ZIKResF3ivijLYopJWfftamaFDyY+463\n' +
        'uI/WNFWrc7iQWI4Fy3eUezuqBZhiZdWMxuP7yDwLgWr2h3CUNYoLvocrRuye6cI8fjmDWzeTdun\n' +
        '5tSM8OL30hY7aH1n/ZDweng0w5k8i7Znm5LpLaAJzkA7CFV0CXI56s3isJ5CjXJFm"}',
    backupKey: '{"iv":"yoGM9gzwedwMp+D5xfZjuw==","v":1,"iter":10000,"ks":256,"ts":64,"mode"\n' +
        ':"ccm","adata":"","cipher":"aes","salt":"bLoJpDYNb1o=","ct":"rK3JZ5VyxqsfgV\n' +
        'JVrR8bddfMe5cy963ntRt3B7l71tvHT8KleNBlzPN6VPrvw1T9KhTHPZ8zDq1pLN39wdrgr4lD4\n' +
        '7+FwsclqQscsi97YW7pTHhWBGGMCKzItXQfeXYCZqvXu/BeV80VWKwso40vZ13pNWmZRhjy/40U\n' +
        'Z0J/Tf+2NuFSEuLKznrgD/oLJaAkYS7d1Bt+GUHWkXnDOsLwqbBL8y5TQJkZrp37nDXXRdP1IE1\n' +
        'MOm2C+GMvboHwm4bpKQRpVURgjyk/7/XWXQ7OUvIFgq324sQbQCHVB/voaRVF+3x7ShRjiGG9sE\n' +
        'W0+XsbbamrXvGiexAGPDM+2kuPe/rBFhJgf4HJnYFAM9sjzWdOIedhkuOokwqgd5OqreVMO/dYF\n' +
        'PGI29cnrov8JpIQw5gArtlgUxOdelg/oCaRUryHCABVPTODbFYygf+3IAgng2M+oEvh26gVFOfz\n' +
        'oPfBoqqXk6/kWpc3zx2byOB+c6Yzhg9sRnF5UobMBl50nklo1kM/24ObmsNCQeVyuZARjQgpbZx\n' +
        'wn5l51knvxmvQXxvwnfsuifje9GkE4Tt+vYu49o3V8vmRaOxcEMnzj5WOC/MOb6QPT4Jx8/J9CB\n' +
        'V4nU+MvgZbDJKMx9kuyJ62PihBy8uSE1JrNgWU3aEkuh9aD5HEoAtyy4tBatRTzOMLSsSr3hii7\n' +
        'pVTBkfEFVzsrHRfIc4FhsnFqSG1yPtxQ68FUeZNM23pCOSJMcbzuP2deEwOCNE5AFIhVZmHg5Yn\n' +
        'CXgcUwAHh+aD+Jv8LMyHq0U6GKkpghu5bptUJcX6oOwUoKD04NWuE6VvK0ZJhSpxxOLeLI4YrwA\n' +
        'tNVo9KAehWPHJfbx4XyBKu+3Z/mIfPJ2rxz7Izdci/BhDKiuHRUynUJ7/8Taxkjqa6RBdFD/avm\n' +
        'Z1a8GyGlxYbnF3OlJUS0Op3lAXBygMyfZqDYbYSfr7ufaIfULAApxXIV0+tyKeF78yAH/k4GIvH\n' +
        '8GzkFu0MSzhWDddeUuKY4Me36GmPHfkzXw+jGt/I6R2vPZKW5RsfMooV7QKUrINtQTZ/4b8jAsr\n' +
        'NJyYZX1wmjwQu3UzqpN+4oPh+WOJUX1Weo5B6c/4e1FmchSiPcI9xCR64MddbFaA/ORdN4p4cgL\n' +
        'B4K3MOqMpoOCMgo/+N9JQr5FTMrcxVSJWRm7QrSGP3go2tywH1UuJOHxhAthdRw=="}',
    bitgoKey: '74372039ccd6c9b2e1e16e3258ff3307b05527ac5a34e1b603b7b74353599c5eb6de9fc5f0f\n' +
        '5f677bfa1f41915d9e1a2e794f8d5d9770d219bf5432adf0415e3',
    walletPassphrase: '#Bondiola1234',
    walletAddress: '5C8nEaqzwMLC2EFjUFckCfPLeFUVWNp3FErQiN8D5u2ZkWyt',
};
exports.unsignedSweepUser = {
    userKey: '97d9da8bf544d07f3c7ae8ee06bb12a63f74478d0c461201ab94421a9c81d5b379553f0332d\n' +
        'aaa85cb8c65e1fe7e6907c62268c00812b08a4e1361178e304ddb',
    backupKey: '97d9da8bf544d07f3c7ae8ee06bb12a63f74478d0c461201ab94421a9c81d5b379553f0332d\n' +
        'aaa85cb8c65e1fe7e6907c62268c00812b08a4e1361178e304ddb',
    bitgoKey: '97d9da8bf544d07f3c7ae8ee06bb12a63f74478d0c461201ab94421a9c81d5b379553f0332d\n' +
        'aaa85cb8c65e1fe7e6907c62268c00812b08a4e1361178e304ddb',
    walletPassphrase: '#Bondiola1234',
    walletAddress: '5DnZQqbxtB3CWkLgfA6wpqdkCd9Bq7AX49RpLPJwW9mzoT7s',
};
exports.testnetBlock = {
    blockNumber: 12640277,
    hash: '0x3771101969cc5cf1b37db0bdce589fe9174a1b820341c4d3673ce301352a3d42',
};
