// @bitgo/sdk-core/dist/src/bitgo/tss/ecdsa/ecdsa.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.secondaryDelimeter = exports.delimeter = void 0;
exports.createCombinedKey = createCombinedKey;
exports.createUserSignShare = createUserSignShare;
exports.createUserGammaAndMuShare = createUserGammaAndMuShare;
exports.createUserOmicronAndDeltaShare = createUserOmicronAndDeltaShare;
exports.createUserSignatureShare = createUserSignatureShare;
exports.sendShareToBitgo = sendShareToBitgo;
exports.getBitgoToUserLatestShare = getBitgoToUserLatestShare;
exports.encryptNShare = encryptNShare;
exports.buildNShareFromAPIKeyShare = buildNShareFromAPIKeyShare;
exports.decryptNShare = decryptNShare;
exports.getPublicKey = getPublicKey;
exports.parseKShare = parseKShare;
exports.convertKShare = convertKShare;
exports.parseAShare = parseAShare;
exports.convertAShare = convertAShare;
exports.parseMuShare = parseMuShare;
exports.convertMuShare = convertMuShare;
exports.parseDShare = parseDShare;
exports.convertDShare = convertDShare;
exports.parseSDShare = parseSDShare;
exports.convertSDShare = convertSDShare;
exports.parseSignatureShare = parseSignatureShare;
exports.convertCombinedSignature = convertCombinedSignature;
exports.parseCombinedSignature = parseCombinedSignature;
exports.convertSignatureShare = convertSignatureShare;
exports.convertBShare = convertBShare;
exports.parseBShare = parseBShare;
exports.convertOShare = convertOShare;
exports.parseOShare = parseOShare;
exports.getParticipantIndex = getParticipantIndex;
exports.getParticipantFromIndex = getParticipantFromIndex;
exports.verifyWalletSignature = verifyWalletSignature;
const tss_1 = require("./../../../account-lib/mpc/tss");
const types_1 = require("./types");
const utils_1 = require("../../utils");
const types_2 = require("../types");
const common_1 = require("../common");
const keccak_1 = __importDefault(require("keccak"));
const assert_1 = __importDefault(require("assert"));
const utxo_lib_1 = require("@bitgo/utxo-lib");
const pgp = __importStar(require("openpgp"));
const bs58_1 = __importDefault(require("bs58"));
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
const MPC = new tss_1.Ecdsa();
/**
 * Combines NShares to combine the final TSS key
 * This can only be used to create the User or Backup key since it requires the common keychain from BitGo first
 *
 * @param keyShare - TSS key share
 * @param encryptedNShares - encrypted NShares with information on how to decrypt
 * @param commonKeychain - expected common keychain of the combined key
 * @returns {CombinedKey} combined TSS key
 */
async function createCombinedKey(keyShare, encryptedNShares, commonKeychain) {
    const nShares = [];
    let bitgoNShare;
    let userNShare;
    let backupNShare;
    for (const encryptedNShare of encryptedNShares) {
        const nShare = await decryptNShare(encryptedNShare, encryptedNShare.isbs58Encoded);
        switch (encryptedNShare.nShare.j) {
            case 1:
                userNShare = nShare;
                break;
            case 2:
                backupNShare = nShare;
                break;
            case 3:
                bitgoNShare = nShare;
                break;
            default:
                throw new Error('Invalid NShare index');
        }
        nShares.push(nShare);
    }
    if (!bitgoNShare) {
        throw new Error('Missing BitGo N Share');
    }
    const combinedKey = MPC.keyCombine(keyShare.pShare, nShares);
    if (combinedKey.xShare.y + combinedKey.xShare.chaincode !== commonKeychain) {
        throw new Error('Common keychains do not match');
    }
    const signingMaterial = {
        pShare: keyShare.pShare,
        bitgoNShare,
        backupNShare,
        userNShare,
    };
    return {
        signingMaterial,
        commonKeychain,
    };
}
/**
 * Creates the SignShare with User XShare and YShare Corresponding to BitGo
 * @param {XShare} xShare User secret xShare
 * @param {YShare} yShare YShare from Bitgo
 * @returns {Promise<SignShare>}
 */
async function createUserSignShare(xShare, yShare) {
    if (xShare.i !== types_2.ShareKeyPosition.USER) {
        throw new Error(`Invalid XShare, XShare doesn't belong to the User`);
    }
    if (yShare.i !== types_2.ShareKeyPosition.USER || yShare.j !== types_2.ShareKeyPosition.BITGO) {
        throw new Error('Invalid YShare provided for sign');
    }
    return await MPC.signShare(xShare, yShare);
}
/**
 * Creates the Gamma Share and MuShare with User WShare and AShare From BitGo
 * @param {WShare} wShare User WShare
 * @param {AShare} aShare AShare from Bitgo
 * @returns {Promise<SignConvertStep2Response>}
 */
async function createUserGammaAndMuShare(wShare, aShare) {
    if (wShare.i !== types_2.ShareKeyPosition.USER) {
        throw new Error(`Invalid WShare, doesn't belong to the User`);
    }
    if (aShare.i !== types_2.ShareKeyPosition.USER || aShare.j !== types_2.ShareKeyPosition.BITGO) {
        throw new Error('Invalid AShare, is not from Bitgo to User');
    }
    return MPC.signConvertStep2({ wShare, aShare });
}
/**
 * Creates the Omicron Share and Delta share with user GShare
 * @param {GShare} gShare User GShare
 * @returns {Promise<CreateUserOmicronAndDeltaShareRT>}
 */
async function createUserOmicronAndDeltaShare(gShare) {
    if (gShare.i !== types_2.ShareKeyPosition.USER) {
        throw new Error(`Invalid GShare, doesn't belong to the User`);
    }
    return MPC.signCombine({
        gShare: gShare,
        signIndex: {
            i: types_2.ShareKeyPosition.BITGO,
            j: gShare.i,
        },
    });
}
/**
 * Creates the Signature Share with User OShare and DShare From BitGo
 * @param {OShare} oShare User OShare
 * @param {DShare} dShare DShare from bitgo
 * @param {Buffer} message message to perform sign
 * @returns {Promise<createUserSignShareRT>}
 */
async function createUserSignatureShare(oShare, dShare, message, hash = (0, keccak_1.default)('keccak256')) {
    if (oShare.i !== types_2.ShareKeyPosition.USER) {
        throw new Error(`Invalid OShare, doesn't belong to the User`);
    }
    if (dShare.i !== types_2.ShareKeyPosition.USER || dShare.j !== types_2.ShareKeyPosition.BITGO) {
        throw new Error(`Invalid DShare, doesn't seem to be from BitGo`);
    }
    return MPC.sign(message, oShare, dShare, hash);
}
/**
 * Sends Share To Bitgo
 * @param {BitGoBase} bitgo - the bitgo instance
 * @param {String} walletId - the wallet id  *
 * @param {String} txRequestId - the txRequest Id
 * @param requestType - the type of request being submitted (either tx or message for signing)
 * @param shareType
 * @param share
 * @param signerShare
 * @param vssProof - the v value of the share
 * @param privateShareProof - the uSig of the share
 * @param publicShare - the y value of the share
 * @param userPublicGpgKey - the public key of the gpg key used for creating the privateShareProof
 * @param reqId - request tracer request id
 * @returns {Promise<SignatureShareRecord>} - a Signature Share
 */
async function sendShareToBitgo(bitgo, walletId, txRequestId, requestType, shareType, share, signerShare, vssProof, privateShareProof, publicShare, userPublicGpgKey, reqId) {
    if (shareType !== types_1.SendShareType.SShare && share.i !== types_2.ShareKeyPosition.BITGO) {
        throw new Error('Invalid Share, is not from User to Bitgo');
    }
    let signatureShare;
    let responseFromBitgo;
    switch (shareType) {
        case types_1.SendShareType.KShare:
            (0, assert_1.default)(signerShare, `signer share must be present`);
            const kShare = share;
            signatureShare = convertKShare(kShare);
            signatureShare.vssProof = vssProof;
            signatureShare.publicShare = publicShare;
            signatureShare.privateShareProof = privateShareProof;
            await (0, common_1.sendSignatureShare)(bitgo, walletId, txRequestId, signatureShare, requestType, signerShare, 'ecdsa', 'full', userPublicGpgKey, reqId);
            responseFromBitgo = await getBitgoToUserLatestShare(bitgo, walletId, txRequestId, types_1.ReceivedShareType.AShare, requestType, reqId);
            break;
        case types_1.SendShareType.MUShare:
            const shareToSend = share;
            const muShareRecord = convertMuShare(shareToSend.muShare);
            const dShareRecord = convertDShare(shareToSend.dShare);
            signatureShare = {
                to: utils_1.SignatureShareType.BITGO,
                from: getParticipantFromIndex(shareToSend.dShare.j),
                share: `${muShareRecord.share}${exports.secondaryDelimeter}${dShareRecord.share}`,
            };
            await (0, common_1.sendSignatureShare)(bitgo, walletId, txRequestId, signatureShare, requestType, signerShare, 'ecdsa', undefined, undefined, reqId);
            responseFromBitgo = await getBitgoToUserLatestShare(bitgo, walletId, txRequestId, types_1.ReceivedShareType.DShare, requestType, reqId);
            break;
        case types_1.SendShareType.SShare:
            const sShare = share;
            signatureShare = convertSignatureShare(sShare, 1, 3);
            await (0, common_1.sendSignatureShare)(bitgo, walletId, txRequestId, signatureShare, requestType, signerShare, 'ecdsa', undefined, undefined, reqId);
            responseFromBitgo = sShare;
            break;
        default:
            throw new Error('Invalid Share given to send');
    }
    return responseFromBitgo;
}
/**
 * Gets the latest user's share from bitgo needed to continue signing flow
 * @param {BitGoBase} bitgo - the bitgo instance
 * @param {String} walletId - the wallet id  *
 * @param {String} txRequestId - the txRequest Id
 * @param {ReceivedShareType} shareType - the excpected share type
 * @param {IRequestTracer} reqId - request tracer request id
 * @returns {Promise<SendShareToBitgoRT>} - share from bitgo to user
 */
async function getBitgoToUserLatestShare(bitgo, walletId, txRequestId, shareType, requestType, reqId) {
    let responseFromBitgo;
    const txRequest = await (0, common_1.getTxRequest)(bitgo, walletId, txRequestId, reqId);
    let userShares;
    switch (requestType) {
        case utils_1.RequestType.tx:
            (0, assert_1.default)(txRequest.transactions, 'transactions required as part of txRequest');
            userShares = txRequest.transactions[0].signatureShares;
            break;
        case utils_1.RequestType.message:
            (0, assert_1.default)(txRequest.messages, 'messages required as part of txRequest');
            userShares = txRequest.messages[0].signatureShares;
            break;
    }
    if (!userShares || !userShares.length) {
        throw new Error('user share is not present');
    }
    const shareRecord = userShares[userShares.length - 1];
    switch (shareType) {
        case types_1.ReceivedShareType.AShare:
            responseFromBitgo = parseAShare(shareRecord);
            break;
        case types_1.ReceivedShareType.DShare:
            responseFromBitgo = parseDShare(shareRecord);
            break;
        case types_1.ReceivedShareType.Signature:
            responseFromBitgo = parseSignatureShare(shareRecord);
            break;
        default:
            throw new Error('Invalid share received');
    }
    return responseFromBitgo;
}
/**
 * Prepares a NShare to be exchanged with other key holders.
 * Output is in a format that is usable within BitGo's ecosystem.
 *
 * @param keyShare - TSS key share of the party preparing exchange materials
 * @param recipientIndex - index of the recipient (1, 2, or 3)
 * @param recipientGpgPublicArmor - recipient's public gpg key in armor format
 * @param senderGpgKey - ephemeral GPG key to encrypt / decrypt sensitve data exchanged between user and server
 * @param isbs58Encoded - is bs58 encoded or not
 * @returns encrypted N Share
 */
async function encryptNShare(keyShare, recipientIndex, recipientGpgPublicArmor, senderGpgKey, isbs58Encoded = true) {
    const nShare = keyShare.nShares[recipientIndex];
    if (!nShare) {
        throw new Error('Invalid recipient');
    }
    const publicShare = Buffer.concat([
        Buffer.from(keyShare.pShare.y, 'hex'),
        Buffer.from(keyShare.pShare.chaincode, 'hex'),
    ]).toString('hex');
    let privateShare;
    if (isbs58Encoded) {
        privateShare = utxo_lib_1.bip32.fromPrivateKey(Buffer.from(nShare.u, 'hex'), Buffer.from(nShare.chaincode, 'hex')).toBase58();
    }
    else {
        privateShare = Buffer.concat([Buffer.from(nShare.u, 'hex'), Buffer.from(nShare.chaincode, 'hex')]).toString('hex');
    }
    const recipientPublicKey = await pgp.readKey({ armoredKey: recipientGpgPublicArmor });
    const encryptedPrivateShare = (await pgp.encrypt({
        message: await pgp.createMessage({
            text: privateShare,
        }),
        encryptionKeys: [recipientPublicKey],
    }));
    return {
        i: nShare.i,
        j: nShare.j,
        publicShare,
        encryptedPrivateShare,
        n: nShare.n,
        vssProof: nShare.v,
        privateShareProof: await (0, utils_1.createShareProof)(senderGpgKey.privateKey, nShare.u, 'ecdsa'),
    };
}
/**
 * Prepares a NShare to be exchanged with other key holders.
 * An API key share received from a third party should already be encrypted
 *
 * @param keyShare - TSS key share of the party preparing exchange materials
 * @returns encrypted N Share
 */
async function buildNShareFromAPIKeyShare(keyShare) {
    return {
        i: getParticipantIndex(keyShare.to),
        j: getParticipantIndex(keyShare.from),
        publicShare: keyShare.publicShare,
        encryptedPrivateShare: keyShare.privateShare,
        n: keyShare.n ?? '', // this is not currently needed for key creation
        privateShareProof: keyShare.privateShareProof,
        vssProof: keyShare.vssProof,
    };
}
/**
 * Decrypts encrypted n share
 * @param encryptedNShare - decryptable n share with recipient private gpg key armor and sender public gpg key
 * @param isbs58Encoded
 * @returns N share
 */
async function decryptNShare(encryptedNShare, isbs58Encoded = true) {
    const recipientPrivateKey = await pgp.readKey({ armoredKey: encryptedNShare.recipientPrivateArmor });
    const prv = (await pgp.decrypt({
        message: await pgp.readMessage({ armoredMessage: encryptedNShare.nShare.encryptedPrivateShare }),
        decryptionKeys: [recipientPrivateKey],
    })).data;
    let u;
    if (isbs58Encoded) {
        const privateShare = bs58_1.default.decode(prv).toString('hex');
        u = privateShare.slice(92, 156);
    }
    else {
        u = prv.slice(0, 64);
    }
    return {
        i: encryptedNShare.nShare.i,
        j: encryptedNShare.nShare.j,
        n: encryptedNShare.nShare.n,
        y: encryptedNShare.nShare.publicShare.slice(0, 66),
        u: u,
        chaincode: encryptedNShare.nShare.publicShare.slice(66, 130),
        v: encryptedNShare.nShare.vssProof,
    };
}
/**
 * Gets public key from common keychain
 * @param commonKeyChain - common keychain of ecdsa tss
 * @returns public key
 */
function getPublicKey(commonKeyChain) {
    return commonKeyChain.slice(0, 66);
}
exports.delimeter = ':';
exports.secondaryDelimeter = '-';
function validateSharesLength(shares, expectedLength, shareName) {
    if (shares.length < expectedLength) {
        throw new Error(`Invalid ${shareName} share`);
    }
}
function validateOptionalValues(shares, start, end, shareName, valueName) {
    let found = false;
    for (let i = start; i < end; i++) {
        if (shares[i]) {
            found = true;
        }
        else if (found) {
            throw new Error(`Inconsistent optional ${valueName} value in ${shareName} share`);
        }
    }
    return found;
}
/**
 * parses K share from signature share record
 * @param share - signature share record
 * @returns K Share
 */
function parseKShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 11 + 2 * sdk_lib_mpc_1.EcdsaPaillierProof.m, 'K');
    const hasProof = validateOptionalValues(shares, 5, 11, 'K', 'proof');
    const proof = hasProof
        ? {
            z: shares[5],
            u: shares[6],
            w: shares[7],
            s: shares[8],
            s1: shares[9],
            s2: shares[10],
        }
        : undefined;
    return {
        i: getParticipantIndex(share.to),
        j: getParticipantIndex(share.from),
        k: shares[0],
        n: shares[1],
        ntilde: shares[2],
        h1: shares[3],
        h2: shares[4],
        proof,
        p: shares.slice(11, 11 + sdk_lib_mpc_1.EcdsaPaillierProof.m),
        sigma: shares.slice(11 + sdk_lib_mpc_1.EcdsaPaillierProof.m, 11 + 2 * sdk_lib_mpc_1.EcdsaPaillierProof.m),
    };
}
/**
 * convert K share to signature share record
 * @param share - K share
 * @returns signature share record
 */
function convertKShare(share) {
    return {
        to: getParticipantFromIndex(share.i),
        from: getParticipantFromIndex(share.j),
        share: `${share.k}${exports.delimeter}${share.n}${exports.delimeter}${share.ntilde}${exports.delimeter}${share.h1}${exports.delimeter}${share.h2}${exports.delimeter}${share.proof?.z || ''}${exports.delimeter}${share.proof?.u || ''}${exports.delimeter}${share.proof?.w || ''}${exports.delimeter}${share.proof?.s || ''}${exports.delimeter}${share.proof?.s1 || ''}${exports.delimeter}${share.proof?.s2 || ''}${exports.delimeter}${(share.p || []).join(exports.delimeter)}${exports.delimeter}${(share.sigma || []).join(exports.delimeter)}`,
    };
}
/**
 * parses A share from signature share record
 * @param share - signature share record
 * @returns A Share
 */
function parseAShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 37 + sdk_lib_mpc_1.EcdsaPaillierProof.m, 'A');
    const hasProof = validateOptionalValues(shares, 7, 13, 'A', 'proof');
    const hasGammaProof = validateOptionalValues(shares, 13, 25, 'A', 'gammaProof');
    const hasWProof = validateOptionalValues(shares, 25, 37, 'A', 'wProof');
    const proof = hasProof
        ? {
            z: shares[7],
            u: shares[8],
            w: shares[9],
            s: shares[10],
            s1: shares[11],
            s2: shares[12],
        }
        : undefined;
    const gammaProof = hasGammaProof
        ? {
            z: shares[13],
            zprm: shares[14],
            t: shares[15],
            v: shares[16],
            w: shares[17],
            s: shares[18],
            s1: shares[19],
            s2: shares[20],
            t1: shares[21],
            t2: shares[22],
            u: shares[23],
            x: shares[24],
        }
        : undefined;
    const wProof = hasWProof
        ? {
            z: shares[25],
            zprm: shares[26],
            t: shares[27],
            v: shares[28],
            w: shares[29],
            s: shares[30],
            s1: shares[31],
            s2: shares[32],
            t1: shares[33],
            t2: shares[34],
            u: shares[35],
            x: shares[36],
        }
        : undefined;
    return {
        i: getParticipantIndex(share.to),
        j: getParticipantIndex(share.from),
        k: shares[0],
        alpha: shares[1],
        mu: shares[2],
        n: shares[3],
        ntilde: shares[4],
        h1: shares[5],
        h2: shares[6],
        proof,
        gammaProof,
        wProof,
        sigma: shares.slice(37),
    };
}
/**
 * convert A share to signature share record
 * @param share - A share
 * @returns signature share record
 */
function convertAShare(share) {
    return {
        to: getParticipantFromIndex(share.i),
        from: getParticipantFromIndex(share.j),
        share: `${share.k}${exports.delimeter}${share.alpha}${exports.delimeter}${share.mu}${exports.delimeter}${share.n}${exports.delimeter}${share.ntilde}${exports.delimeter}${share.h1}${exports.delimeter}${share.h2}${exports.delimeter}${share.proof?.z || ''}${exports.delimeter}${share.proof?.u || ''}${exports.delimeter}${share.proof?.w || ''}${exports.delimeter}${share.proof?.s || ''}${exports.delimeter}${share.proof?.s1 || ''}${exports.delimeter}${share.proof?.s2 || ''}${exports.delimeter}${share.gammaProof?.z || ''}${exports.delimeter}${share.gammaProof?.zprm || ''}${exports.delimeter}${share.gammaProof?.t || ''}${exports.delimeter}${share.gammaProof?.v || ''}${exports.delimeter}${share.gammaProof?.w || ''}${exports.delimeter}${share.gammaProof?.s || ''}${exports.delimeter}${share.gammaProof?.s1 || ''}${exports.delimeter}${share.gammaProof?.s2 || ''}${exports.delimeter}${share.gammaProof?.t1 || ''}${exports.delimeter}${share.gammaProof?.t2 || ''}${exports.delimeter}${share.gammaProof?.u || ''}${exports.delimeter}${share.gammaProof?.x || ''}${exports.delimeter}${share.wProof?.z || ''}${exports.delimeter}${share.wProof?.zprm || ''}${exports.delimeter}${share.wProof?.t || ''}${exports.delimeter}${share.wProof?.v || ''}${exports.delimeter}${share.wProof?.w || ''}${exports.delimeter}${share.wProof?.s || ''}${exports.delimeter}${share.wProof?.s1 || ''}${exports.delimeter}${share.wProof?.s2 || ''}${exports.delimeter}${share.wProof?.t1 || ''}${exports.delimeter}${share.wProof?.t2 || ''}${exports.delimeter}${share.wProof?.u || ''}${exports.delimeter}${share.wProof?.x || ''}${exports.delimeter}${(share.sigma || []).join(exports.delimeter)}`,
    };
}
/**
 * parses Mu share from signature share record
 * @param share - signature share record
 * @returns Mu Share
 */
function parseMuShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 26, 'Mu');
    const hasGammaProof = validateOptionalValues(shares, 2, 14, 'Mu', 'gammaProof');
    const hasWProof = validateOptionalValues(shares, 14, 26, 'Mu', 'wProof');
    let gammaProof;
    if (hasGammaProof) {
        gammaProof = {
            z: shares[2],
            zprm: shares[3],
            t: shares[4],
            v: shares[5],
            w: shares[6],
            s: shares[7],
            s1: shares[8],
            s2: shares[9],
            t1: shares[10],
            t2: shares[11],
            u: shares[12],
            x: shares[13],
        };
    }
    let wProof;
    if (hasWProof) {
        wProof = {
            z: shares[14],
            zprm: shares[15],
            t: shares[16],
            v: shares[17],
            w: shares[18],
            s: shares[19],
            s1: shares[20],
            s2: shares[21],
            t1: shares[22],
            t2: shares[23],
            u: shares[24],
            x: shares[25],
        };
    }
    return {
        i: getParticipantIndex(share.to),
        j: getParticipantIndex(share.from),
        alpha: shares[0],
        mu: shares[1],
        gammaProof,
        wProof,
    };
}
/**
 * convert Mu share to signature share record
 * @param share - Mu share
 * @returns signature share record
 */
function convertMuShare(share) {
    return {
        to: getParticipantFromIndex(share.i),
        from: getParticipantFromIndex(share.j),
        share: `${share.alpha}${exports.delimeter}${share.mu}${exports.delimeter}${share.gammaProof?.z || ''}${exports.delimeter}${share.gammaProof?.zprm || ''}${exports.delimeter}${share.gammaProof?.t || ''}${exports.delimeter}${share.gammaProof?.v || ''}${exports.delimeter}${share.gammaProof?.w || ''}${exports.delimeter}${share.gammaProof?.s || ''}${exports.delimeter}${share.gammaProof?.s1 || ''}${exports.delimeter}${share.gammaProof?.s2 || ''}${exports.delimeter}${share.gammaProof?.t1 || ''}${exports.delimeter}${share.gammaProof?.t2 || ''}${exports.delimeter}${share.gammaProof?.u || ''}${exports.delimeter}${share.gammaProof?.x || ''}${exports.delimeter}${share.wProof?.z || ''}${exports.delimeter}${share.wProof?.zprm || ''}${exports.delimeter}${share.wProof?.t || ''}${exports.delimeter}${share.wProof?.v || ''}${exports.delimeter}${share.wProof?.w || ''}${exports.delimeter}${share.wProof?.s || ''}${exports.delimeter}${share.wProof?.s1 || ''}${exports.delimeter}${share.wProof?.s2 || ''}${exports.delimeter}${share.wProof?.t1 || ''}${exports.delimeter}${share.wProof?.t2 || ''}${exports.delimeter}${share.wProof?.u || ''}${exports.delimeter}${share.wProof?.x || ''}`,
    };
}
/**
 * parses D share from signature share record
 * @param share - signature share record
 * @returns D Share
 */
function parseDShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 2, 'D');
    return {
        i: getParticipantIndex(share.to),
        j: getParticipantIndex(share.from),
        delta: shares[0],
        Gamma: shares[1],
    };
}
/**
 * convert D share to signature share record
 * @param share - D share
 * @returns signature share record
 */
function convertDShare(share) {
    return {
        to: getParticipantFromIndex(share.i),
        from: getParticipantFromIndex(share.j),
        share: `${share.delta}${exports.delimeter}${share.Gamma}`,
    };
}
/**
 * parses S and D share from signature share record
 * @param share - signature share record
 * @returns Object containing S and D Share
 */
function parseSDShare(share) {
    const shares = share.share.split(exports.secondaryDelimeter);
    validateSharesLength(shares, 2, 'SD');
    return {
        sShare: parseSignatureShare({ to: share.to, from: share.from, share: shares[0] }),
        dShare: parseDShare({ to: share.to, from: share.from, share: shares[1] }),
    };
}
/**
 * convert S and D share to signature share record
 * @param share - S and D share in a object
 * @returns signature share record
 */
function convertSDShare(share) {
    return {
        to: getParticipantFromIndex(share.dShare.i),
        from: getParticipantFromIndex(share.dShare.j),
        share: `${share.sShare.R}${exports.delimeter}${share.sShare.s}${exports.delimeter}${share.sShare.y}${exports.secondaryDelimeter}${share.dShare.delta}${exports.delimeter}${share.dShare.Gamma}`,
    };
}
/**
 * parses signature share from signature share record
 * @param share - signature share record
 * @returns Signature Share
 */
function parseSignatureShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 3, 'Signature');
    return {
        i: getParticipantIndex(share.to),
        R: shares[0],
        s: shares[1],
        y: shares[2],
    };
}
/**
 * convets combined signature to signature share record
 * @param signature - combined signature share
 * @param userIndex - user index, either 1 (user) or 2 (backup)
 * @returns signature share record
 */
function convertCombinedSignature(signature, userIndex) {
    return {
        to: utils_1.SignatureShareType.BITGO,
        from: getParticipantFromIndex(userIndex),
        share: `${signature.recid}${exports.delimeter}${signature.r}${exports.delimeter}${signature.s}${exports.delimeter}${signature.y}`,
    };
}
function parseCombinedSignature(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 3, 'Signature');
    return {
        recid: Number(shares[0]),
        r: shares[1],
        s: shares[2],
        y: shares[3],
    };
}
/**
 * convert signature share to signature share record
 * @param share - Signature share
 * @param senderIndex
 * @param recipientIndex
 * @returns signature share record
 */
function convertSignatureShare(share, senderIndex, recipientIndex) {
    return {
        to: getParticipantFromIndex(recipientIndex),
        from: getParticipantFromIndex(senderIndex),
        share: `${share.R}${exports.delimeter}${share.s}${exports.delimeter}${share.y}`,
    };
}
/**
 * converts B share to signature share record
 * @param share - B share
 * @returns signature share record
 */
function convertBShare(share) {
    return {
        to: utils_1.SignatureShareType.BITGO,
        from: getParticipantFromIndex(share.i),
        share: `${share.beta}${exports.delimeter}${share.gamma}${exports.delimeter}${share.k}${exports.delimeter}${share.nu}${exports.delimeter}${share.w}${exports.delimeter}${share.y}${exports.delimeter}${share.l}${exports.delimeter}${share.m}${exports.delimeter}${share.n}${exports.delimeter}${share.ntilde}${exports.delimeter}${share.h1}${exports.delimeter}${share.h2}${exports.delimeter}${share.ck}${exports.delimeter}${(share.p || []).join(exports.delimeter)}`,
    };
}
/**
 * parses B share from signature share record
 * @param share B share record
 * @returns B Share
 */
function parseBShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 13 + sdk_lib_mpc_1.EcdsaPaillierProof.m, 'B');
    return {
        i: getParticipantIndex(share.to),
        beta: shares[0],
        gamma: shares[1],
        k: shares[2],
        nu: shares[3],
        w: shares[4],
        y: shares[5],
        l: shares[6],
        m: shares[7],
        n: shares[8],
        ntilde: shares[9],
        h1: shares[10],
        h2: shares[11],
        ck: shares[12],
        p: shares.slice(13, 13 + sdk_lib_mpc_1.EcdsaPaillierProof.m),
    };
}
/**
 * converts O share to signature share record
 * @param share O share
 * @returns signature share record
 */
function convertOShare(share) {
    return {
        to: utils_1.SignatureShareType.BITGO,
        from: getParticipantFromIndex(share.i),
        share: `${share.Gamma}${exports.delimeter}${share.delta}${exports.delimeter}${share.k}${exports.delimeter}${share.omicron}${exports.delimeter}${share.y}`,
    };
}
/**
 * parses O share from signature share record
 * @param share O share record
 * @returns O Share
 */
function parseOShare(share) {
    const shares = share.share.split(exports.delimeter);
    validateSharesLength(shares, 5, 'O');
    return {
        i: getParticipantIndex(share.to),
        Gamma: shares[0],
        delta: shares[1],
        k: shares[2],
        omicron: shares[3],
        y: shares[4],
    };
}
/**
 * gets participant index
 * @param participant - participants (user, backup, or bitgo)
 * @returns index (1, 2, 0r 3)
 */
function getParticipantIndex(participant) {
    switch (participant) {
        case 'user':
            return 1;
        case 'backup':
            return 2;
        case 'bitgo':
            return 3;
        default:
            throw Error('Unkown participant');
    }
}
/**
 * gets participant name by index
 * @param index participant index
 * @returns participant name
 */
function getParticipantFromIndex(index) {
    switch (index) {
        case 1:
            return utils_1.SignatureShareType.USER;
        case 2:
            return utils_1.SignatureShareType.BACKUP;
        case 3:
            return utils_1.SignatureShareType.BITGO;
        default:
            throw new Error(`Unknown participant index ${index}`);
    }
}
/**
 * Helper function to verify u-value wallet signatures for the bitgo-user and bitgo-backup shares.
 * @param params
 */
async function verifyWalletSignature(params) {
    const rawNotations = await (0, common_1.commonVerifyWalletSignature)(params);
    const publicUValueRawNotationIndex = 2 + params.verifierIndex;
    // Derive public form of u-value
    const publicUValue = utxo_lib_1.ecc.pointFromScalar(Buffer.from(params.decryptedShare.slice(0, 64), 'hex'), true);
    (0, assert_1.default)(publicUValue !== null, 'null public u-value');
    // Verify that the u value + chaincode is equal to the proof retrieved from the raw notations
    (0, assert_1.default)(Buffer.from(publicUValue).toString('hex') + params.decryptedShare.slice(64) ===
        Buffer.from(rawNotations[publicUValueRawNotationIndex].value).toString(), 'bitgo share mismatch');
}
