// @bitgo/utxo-core/dist/cjs/test/paygo/psbt/payGoAddressProof.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sig = exports.addressProofMsgBuffer = exports.addressProofEntropy = exports.addressToVerify = exports.attestationPrvKey = exports.attestationPubKey = exports.network = void 0;
const assert_1 = __importDefault(require("assert"));
const crypto_1 = __importDefault(require("crypto"));
const secp256k1_1 = require("@bitgo/secp256k1");
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const proprietaryKeyVal_1 = require("bip174/src/lib/proprietaryKeyVal");
const utils_1 = require("bip174/src/lib/utils");
const payGoAddressProof_1 = require("../../../src/paygo/psbt/payGoAddressProof");
const bip32utils_1 = require("../../../src/bip32utils");
const attestation_1 = require("../../../src/paygo/attestation");
// To construct our PSBTs
exports.network = utxolib.networks.bitcoin;
const keys = [1, 2, 3].map((v) => secp256k1_1.bip32.fromSeed(Buffer.alloc(16, `test/2/${v}`), exports.network));
const rootWalletKeys = new utxolib.bitgo.RootWalletKeys([keys[0], keys[1], keys[2]]);
// PSBT INPUTS AND OUTPUTS
const psbtInputs = utxolib.testutil.inputScriptTypes.map((scriptType) => ({
    scriptType,
    value: BigInt(1000),
}));
const psbtOutputs = utxolib.testutil.outputScriptTypes.map((scriptType) => ({
    scriptType,
    value: BigInt(900),
}));
// wallet pub and priv key for tbtc
const dummyPub1 = rootWalletKeys.deriveForChainAndIndex(50, 200);
exports.attestationPubKey = dummyPub1.user.publicKey;
exports.attestationPrvKey = dummyPub1.user.privateKey;
// our xpub converted to base58 address
exports.addressToVerify = utxolib.address.toBase58Check(utxolib.crypto.hash160(Buffer.from(dummyPub1.backup.publicKey)), utxolib.networks.bitcoin.pubKeyHash, utxolib.networks.bitcoin);
// Fixed entropy for deterministic test fixtures (64 bytes of 0x00)
exports.addressProofEntropy = Buffer.alloc(64, 0x00);
// Create the message buffer manually with our fixed entropy
const addressBuffer = Buffer.from(exports.addressToVerify);
const uuidBuffer = Buffer.from(attestation_1.NIL_UUID);
exports.addressProofMsgBuffer = Buffer.concat([exports.addressProofEntropy, addressBuffer, uuidBuffer]);
// signature with the given msg addressProofMsgBuffer
exports.sig = (0, bip32utils_1.signMessage)(exports.addressProofMsgBuffer, exports.attestationPrvKey, exports.network);
function getTestPsbt() {
    return utxolib.testutil.constructPsbt(psbtInputs, psbtOutputs, exports.network, rootWalletKeys, 'unsigned');
}
describe('test fixtures', () => {
    it('should have expected test fixture values', () => {
        // Address fixture - derived from backup key
        assert_1.default.strictEqual(exports.addressToVerify, '1CdWUVacSQQJ617HfuNWByGiisEGXGNx2c');
        // Entropy fixture - fixed 64 bytes of zeros for deterministic testing
        assert_1.default.strictEqual(exports.addressProofEntropy.toString('hex'), '0000000000000000000000000000000000000000000000000000000000000000' +
            '0000000000000000000000000000000000000000000000000000000000000000');
        assert_1.default.strictEqual(exports.addressProofEntropy.length, 64);
        // Signature fixture - Bitcoin message signature (65 bytes with recovery ID)
        assert_1.default.strictEqual(exports.sig.toString('hex'), '1fd62abac20bb963f5150aa4b3f4753c5f2f53ced5183ab7761d0c95c2820f6b' +
            'b722b6d0d9adbab782d2d0d66402794b6bd6449dc26f634035ee388a2b5e7b53f6');
        assert_1.default.strictEqual(exports.sig.length, 65);
        // Public key fixture - compressed secp256k1 public key (33 bytes)
        assert_1.default.strictEqual(exports.attestationPubKey.toString('hex'), '02456f4f788b6af55eb9c54d88692cadef4babdbc34cde75218cc1d6b6de3dea2d');
        assert_1.default.strictEqual(exports.attestationPubKey.length, 33);
        // Message buffer structure: [ENTROPY][ADDRESS][UUID]
        // Total length: 64 (entropy) + 34 (address) + 36 (UUID) = 134
        assert_1.default.strictEqual(exports.addressProofMsgBuffer.length, 134);
        // Verify message components
        const entropy = exports.addressProofMsgBuffer.subarray(0, 64);
        const address = exports.addressProofMsgBuffer.subarray(64, 98);
        const uuid = exports.addressProofMsgBuffer.subarray(98, 134);
        assert_1.default.strictEqual(entropy.toString('hex'), exports.addressProofEntropy.toString('hex'));
        assert_1.default.strictEqual(address.toString(), exports.addressToVerify);
        assert_1.default.strictEqual(uuid.toString(), attestation_1.NIL_UUID);
    });
});
describe('addPaygoAddressProof and verifyPaygoAddressProof', () => {
    function getPayGoProprietaryKey(proprietaryKeyVals) {
        return proprietaryKeyVals
            .map(({ key, value }) => {
            return { key: (0, proprietaryKeyVal_1.decodeProprietaryKey)(key), value };
        })
            .filter((keyValue) => {
            return (keyValue.key.identifier === utxolib.bitgo.PSBT_PROPRIETARY_IDENTIFIER &&
                keyValue.key.subtype === utxolib.bitgo.ProprietaryKeySubtype.PAYGO_ADDRESS_ATTESTATION_PROOF);
        });
    }
    it('should add and verify a valid paygo address proof on the PSBT', () => {
        const psbt = getTestPsbt();
        psbt.addOutput({ script: utxolib.address.toOutputScript(exports.addressToVerify, exports.network), value: BigInt(10000) });
        const outputIndex = psbt.data.outputs.length - 1;
        (0, payGoAddressProof_1.addPayGoAddressProof)(psbt, outputIndex, exports.sig, exports.addressProofEntropy);
        (0, payGoAddressProof_1.verifyPayGoAddressProof)(psbt, outputIndex, exports.attestationPubKey);
    });
    it('should throw an error if there are multiple PayGo proprietary keys in the PSBT', () => {
        const outputIndex = 0;
        const psbt = getTestPsbt();
        (0, payGoAddressProof_1.addPayGoAddressProof)(psbt, outputIndex, exports.sig, exports.addressProofEntropy);
        (0, payGoAddressProof_1.addPayGoAddressProof)(psbt, outputIndex, Buffer.from('signature2'), crypto_1.default.randomBytes(64));
        const output = (0, utils_1.checkForOutput)(psbt.data.outputs, outputIndex);
        const proofInPsbt = getPayGoProprietaryKey(output.unknownKeyVals);
        (0, assert_1.default)(proofInPsbt.length !== 0);
        (0, assert_1.default)(proofInPsbt.length > 1);
        assert_1.default.throws(() => (0, payGoAddressProof_1.verifyPayGoAddressProof)(psbt, outputIndex, exports.attestationPubKey), (e) => e.message === 'There are multiple paygo address proofs encoded in the PSBT. Something went wrong.');
    });
});
describe('verifyPaygoAddressProof', () => {
    it('should throw an error if there is no PayGo address in PSBT', () => {
        const psbt = getTestPsbt();
        assert_1.default.throws(() => (0, payGoAddressProof_1.verifyPayGoAddressProof)(psbt, 0, exports.attestationPubKey), (e) => e.message === 'There is no paygo address proof encoded in the PSBT at output 0.');
    });
});
describe('getPaygoAddressProofIndex', () => {
    it('should get PayGo address proof index from PSBT if there is one', () => {
        const psbt = getTestPsbt();
        const outputIndex = 0;
        (0, payGoAddressProof_1.addPayGoAddressProof)(psbt, outputIndex, exports.sig, Buffer.from(exports.attestationPubKey));
        (0, assert_1.default)((0, payGoAddressProof_1.psbtOutputIncludesPaygoAddressProof)(psbt));
        (0, assert_1.default)((0, payGoAddressProof_1.getPayGoAddressProofOutputIndex)(psbt) === 0);
    });
    it('should return undefined if there is no PayGo address proof in PSBT', () => {
        const psbt = getTestPsbt();
        (0, assert_1.default)((0, payGoAddressProof_1.getPayGoAddressProofOutputIndex)(psbt) === undefined);
        (0, assert_1.default)(!(0, payGoAddressProof_1.psbtOutputIncludesPaygoAddressProof)(psbt));
    });
    it('should return an error and fail if we have multiple PayGo address in the PSBT in the same output index', () => {
        const psbt = getTestPsbt();
        const outputIndex = 0;
        (0, payGoAddressProof_1.addPayGoAddressProof)(psbt, outputIndex, exports.sig, exports.addressProofEntropy);
        (0, payGoAddressProof_1.addPayGoAddressProof)(psbt, outputIndex, exports.sig, crypto_1.default.randomBytes(64));
        assert_1.default.throws(() => (0, payGoAddressProof_1.getPayGoAddressProofOutputIndex)(psbt), (e) => e.message === 'There are multiple PayGo addresses in the PSBT output 0.');
    });
});
