// @bitgo/sdk-coin-avaxp/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INPUT_SEPARATOR = exports.ADDRESS_SEPARATOR = exports.SECP256K1_STAKEABLE_LOCK_OUT = exports.SECP256K1_Transfer_Output = exports.MethodNames = void 0;
/**
 * Method names for the transaction method. Names change based on the type of transaction e.g 'bond' for the staking transaction
 */
var MethodNames;
(function (MethodNames) {
    MethodNames[MethodNames["addDelegator"] = 0] = "addDelegator";
    MethodNames[MethodNames["addValidator"] = 1] = "addValidator";
})(MethodNames || (exports.MethodNames = MethodNames = {}));
/**
 * TypeId value for SECP256K1 Transfer Output
 *
 * {@link https://build.avax.network/docs/api-reference/p-chain/txn-format#secp256k1-transfer-output }
 */
exports.SECP256K1_Transfer_Output = 7;
/**
 * TypeId value for Stakeable Lock Output
 * {@link https://build.avax.network/docs/api-reference/p-chain/txn-format#stakeablelockout }
 */
exports.SECP256K1_STAKEABLE_LOCK_OUT = 22;
exports.ADDRESS_SEPARATOR = '~';
exports.INPUT_SEPARATOR = ':';
