// @bitgo/sdk-coin-avaxp/dist/src/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.TransactionBuilderFactory = exports.DeprecatedTransactionBuilder = exports.DeprecatedTransaction = exports.TransactionBuilder = exports.Transaction = exports.KeyPair = void 0;
const utils_1 = __importDefault(require("./utils"));
exports.Utils = utils_1.default;
__exportStar(require("./iface"), exports);
var keyPair_1 = require("./keyPair");
Object.defineProperty(exports, "KeyPair", { enumerable: true, get: function () { return keyPair_1.KeyPair; } });
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return transaction_1.Transaction; } });
var transactionBuilder_1 = require("./transactionBuilder");
Object.defineProperty(exports, "TransactionBuilder", { enumerable: true, get: function () { return transactionBuilder_1.TransactionBuilder; } });
var deprecatedTransaction_1 = require("./deprecatedTransaction");
Object.defineProperty(exports, "DeprecatedTransaction", { enumerable: true, get: function () { return deprecatedTransaction_1.DeprecatedTransaction; } });
var deprecatedTransactionBuilder_1 = require("./deprecatedTransactionBuilder");
Object.defineProperty(exports, "DeprecatedTransactionBuilder", { enumerable: true, get: function () { return deprecatedTransactionBuilder_1.DeprecatedTransactionBuilder; } });
var transactionBuilderFactory_1 = require("./transactionBuilderFactory");
Object.defineProperty(exports, "TransactionBuilderFactory", { enumerable: true, get: function () { return transactionBuilderFactory_1.TransactionBuilderFactory; } });
__exportStar(require("./iface"), exports);
