// @bitgo/sdk-coin-flrp/dist/test/unit/flrp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const FlrpLib = __importStar(require("../../src/lib"));
const sdk_test_1 = require("@bitgo/sdk-test");
const src_1 = require("../../src/");
const crypto_1 = require("crypto");
const sdk_api_1 = require("@bitgo/sdk-api");
const account_1 = require("../resources/account");
const exportInC_1 = require("../resources/transactionData/exportInC");
const exportInP_1 = require("../resources/transactionData/exportInP");
const importInP_1 = require("../resources/transactionData/importInP");
const importInC_1 = require("../resources/transactionData/importInC");
const sdk_core_1 = require("@bitgo/sdk-core");
const assert_1 = __importDefault(require("assert"));
describe('Flrp test cases', function () {
    const coinName = 'flrp';
    const tcoinName = 't' + coinName;
    let bitgo;
    let basecoin;
    const keychains = [{ pub: account_1.SEED_ACCOUNT.publicKey }, { pub: account_1.ACCOUNT_1.publicKey }, { pub: account_1.ACCOUNT_2.publicKey }];
    before(function () {
        bitgo = sdk_test_1.TestBitGo.decorate(sdk_api_1.BitGoAPI, {
            env: 'mock',
        });
        bitgo.initializeTestVars();
        bitgo.safeRegister(coinName, src_1.Flrp.createInstance);
        bitgo.safeRegister(tcoinName, src_1.TflrP.createInstance);
        basecoin = bitgo.coin(tcoinName);
    });
    it('should instantiate the coin', function () {
        let localBasecoin = bitgo.coin(tcoinName);
        localBasecoin.should.be.an.instanceof(src_1.TflrP);
        localBasecoin = bitgo.coin(coinName);
        localBasecoin.should.be.an.instanceof(src_1.Flrp);
    });
    it('should return ' + tcoinName, function () {
        basecoin.getChain().should.equal(tcoinName);
    });
    it('should return full name', function () {
        basecoin.getFullName().should.equal('Testnet Flare P-Chain');
    });
    it('should return base factor', function () {
        basecoin.getBaseFactor().should.equal(1e9);
    });
    it('should return coin family', function () {
        basecoin.getFamily().should.equal('flrp');
    });
    it('should return default multisig type', function () {
        basecoin.getDefaultMultisigType().should.equal('onchain');
    });
    describe('Keypairs:', () => {
        it('should generate a keypair from random seed', function () {
            const keyPair = basecoin.generateKeyPair();
            keyPair.should.have.property('pub');
            keyPair.should.have.property('prv');
        });
        it('should generate a keypair from a seed', function () {
            const seed = Buffer.from(account_1.SEED_ACCOUNT.seed, 'hex');
            const keyPair = basecoin.generateKeyPair(seed);
            keyPair.pub.should.equal(account_1.SEED_ACCOUNT.publicKey);
            keyPair.prv.should.equal(account_1.SEED_ACCOUNT.privateKey);
        });
        it('should validate a public key', function () {
            basecoin.isValidPub(account_1.SEED_ACCOUNT.publicKey).should.equal(true);
            basecoin.isValidPub(account_1.ACCOUNT_1.publicKey).should.equal(true);
        });
        it('should fail to validate an invalid public key', function () {
            basecoin.isValidPub('invalid').should.equal(false);
        });
        it('should validate a private key', function () {
            basecoin.isValidPrv(account_1.SEED_ACCOUNT.privateKey).should.equal(true);
            basecoin.isValidPrv(account_1.ACCOUNT_1.privateKey).should.equal(true);
        });
        it('should fail to validate an invalid private key', function () {
            basecoin.isValidPrv('invalid').should.equal(false);
        });
    });
    describe('Sign Transaction', () => {
        it('should sign an export from C-chain transaction', async () => {
            const params = {
                txPrebuild: {
                    txHex: exportInC_1.EXPORT_IN_C.unsignedHex,
                },
                prv: exportInC_1.EXPORT_IN_C.privateKey,
            };
            const signedTx = await basecoin.signTransaction(params);
            signedTx.should.have.property('halfSigned');
            const halfSigned = signedTx.halfSigned;
            (0, assert_1.default)(halfSigned, 'halfSigned should be defined');
            (0, assert_1.default)(halfSigned.txHex, 'txHex should be defined');
            halfSigned.txHex.should.equal(exportInC_1.EXPORT_IN_C.signedHex);
        });
        it('should sign an export from P-chain transaction', async () => {
            // privateKeys[2] corresponds to the first signature slot (sorted address order: 3329be7d... is slot 1)
            const params = {
                txPrebuild: {
                    txHex: exportInP_1.EXPORT_IN_P.unsignedHex,
                },
                prv: exportInP_1.EXPORT_IN_P.privateKeys[2],
            };
            const signedTx = await basecoin.signTransaction(params);
            signedTx.should.have.property('halfSigned');
            const halfSigned = signedTx.halfSigned;
            (0, assert_1.default)(halfSigned, 'halfSigned should be defined');
            (0, assert_1.default)(halfSigned.txHex, 'txHex should be defined');
            halfSigned.txHex.should.equal(exportInP_1.EXPORT_IN_P.halfSigntxHex);
        });
        it('should sign an import to P-chain transaction', async () => {
            // privateKeys[2] corresponds to the first signature slot (sorted address order: 3329be7d... is slot 1)
            const params = {
                txPrebuild: {
                    txHex: importInP_1.IMPORT_IN_P.unsignedHex,
                },
                prv: importInP_1.IMPORT_IN_P.privateKeys[2],
            };
            const signedTx = await basecoin.signTransaction(params);
            signedTx.should.have.property('halfSigned');
            const halfSigned = signedTx.halfSigned;
            (0, assert_1.default)(halfSigned, 'halfSigned should be defined');
            (0, assert_1.default)(halfSigned.txHex, 'txHex should be defined');
            halfSigned.txHex.should.equal(importInP_1.IMPORT_IN_P.halfSigntxHex);
        });
        it('should sign an import to C-chain transaction', async () => {
            // privateKeys[2] corresponds to the first signature slot (sorted address order in UTXOs)
            const params = {
                txPrebuild: {
                    txHex: importInC_1.IMPORT_IN_C.unsignedHex,
                },
                prv: importInC_1.IMPORT_IN_C.privateKeys[2],
            };
            const signedTx = await basecoin.signTransaction(params);
            signedTx.should.have.property('halfSigned');
            const halfSigned = signedTx.halfSigned;
            (0, assert_1.default)(halfSigned, 'halfSigned should be defined');
            (0, assert_1.default)(halfSigned.txHex, 'txHex should be defined');
            halfSigned.txHex.should.equal(importInC_1.IMPORT_IN_C.halfSigntxHex);
        });
        it('should reject signing with an invalid key', async () => {
            const params = {
                txPrebuild: {
                    txHex: exportInC_1.EXPORT_IN_C.unsignedHex,
                },
                prv: 'invalid-key',
            };
            await basecoin.signTransaction(params).should.be.rejected();
        });
    });
    describe('Sign Message', () => {
        it('should sign a message', async () => {
            const keyPair = new FlrpLib.KeyPair({ prv: account_1.SEED_ACCOUNT.privateKey });
            const keys = keyPair.getKeys();
            const messageToSign = Buffer.from(account_1.SEED_ACCOUNT.message, 'utf8');
            const signature = await basecoin.signMessage(keys, messageToSign.toString('hex'));
            signature.should.be.instanceOf(Buffer);
            signature.length.should.equal(65);
        });
        it('should sign a random message', async () => {
            const keyPair = new FlrpLib.KeyPair();
            const pubKey = keyPair.getKeys().pub;
            const keys = keyPair.getKeys();
            const messageToSign = Buffer.from((0, crypto_1.randomBytes)(32));
            const signature = await basecoin.signMessage(keys, messageToSign.toString('hex'));
            const verify = FlrpLib.Utils.verifySignature(FlrpLib.Utils.sha256(messageToSign), signature.slice(0, 64), // Remove recovery byte for verification
            Buffer.from(pubKey, 'hex'));
            verify.should.be.true();
        });
        it('should fail to sign with missing private key', async () => {
            const keyPair = new FlrpLib.KeyPair({ pub: account_1.SEED_ACCOUNT.publicKey });
            const keys = keyPair.getKeys();
            const messageToSign = Buffer.from(account_1.SEED_ACCOUNT.message, 'utf8');
            await basecoin
                .signMessage(keys, messageToSign.toString('hex'))
                .should.be.rejectedWith('Invalid key pair options');
        });
    });
    describe('Explain Transaction', () => {
        it('should explain a half signed export from P-chain transaction', async () => {
            const txExplain = await basecoin.explainTransaction({
                halfSigned: { txHex: exportInP_1.EXPORT_IN_P.halfSigntxHex },
            });
            txExplain.type.should.equal(sdk_core_1.TransactionType.Export);
            txExplain.fee.should.have.property('fee');
            txExplain.inputs.should.be.an.Array();
            txExplain.changeAmount.should.equal('14339520'); // 0xDACDC0 from transaction
            txExplain.changeOutputs.should.be.an.Array();
            txExplain.changeOutputs[0].address.should.equal('P-costwo106gc5h5qswhye8e0pmthq4wzf0ekv5qppsrvpu~P-costwo1cueygd7fd37g56s49k3rshqakhp6k8u3adzt6m~P-costwo1xv5mulgpe5lt4tnx2ntnylwe79azu9vpja6lut');
        });
        it('should explain a signed export from P-chain transaction', async () => {
            const txExplain = await basecoin.explainTransaction({ txHex: exportInP_1.EXPORT_IN_P.fullSigntxHex });
            txExplain.type.should.equal(sdk_core_1.TransactionType.Export);
            txExplain.id.should.equal(exportInP_1.EXPORT_IN_P.txhash);
            txExplain.fee.should.have.property('fee');
            txExplain.inputs.should.be.an.Array();
            txExplain.changeAmount.should.equal('14339520'); // 0xDACDC0 from transaction
            txExplain.changeOutputs.should.be.an.Array();
            txExplain.changeOutputs[0].address.should.equal('P-costwo106gc5h5qswhye8e0pmthq4wzf0ekv5qppsrvpu~P-costwo1cueygd7fd37g56s49k3rshqakhp6k8u3adzt6m~P-costwo1xv5mulgpe5lt4tnx2ntnylwe79azu9vpja6lut');
        });
        it('should explain a half signed import to P-chain transaction', async () => {
            const txExplain = await basecoin.explainTransaction({
                halfSigned: { txHex: importInP_1.IMPORT_IN_P.halfSigntxHex },
            });
            txExplain.type.should.equal(sdk_core_1.TransactionType.Import);
            txExplain.fee.should.have.property('fee');
            txExplain.inputs.should.be.an.Array();
            txExplain.outputAmount.should.equal('48739000');
            txExplain.outputs.should.be.an.Array();
            txExplain.outputs.length.should.equal(1);
            txExplain.changeOutputs.should.be.empty();
        });
        it('should explain a signed import to P-chain transaction', async () => {
            const txExplain = await basecoin.explainTransaction({ txHex: importInP_1.IMPORT_IN_P.signedHex });
            txExplain.type.should.equal(sdk_core_1.TransactionType.Import);
            txExplain.id.should.equal(importInP_1.IMPORT_IN_P.txhash);
            txExplain.fee.should.have.property('fee');
            txExplain.inputs.should.be.an.Array();
            txExplain.outputAmount.should.equal('48739000');
            txExplain.outputs.should.be.an.Array();
            txExplain.outputs.length.should.equal(1);
            txExplain.changeOutputs.should.be.empty();
        });
        it('should fail when transaction hex is not provided', async () => {
            await basecoin.explainTransaction({}).should.be.rejectedWith('missing transaction hex');
        });
        it('should fail for invalid transaction hex', async () => {
            await basecoin.explainTransaction({ txHex: 'invalid' }).should.be.rejected();
        });
    });
    describe('Verify Transaction', () => {
        it('should verify an export from C-chain transaction', async () => {
            const txPrebuild = {
                txHex: exportInC_1.EXPORT_IN_C.signedHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: '',
                        amount: exportInC_1.EXPORT_IN_C.amount,
                    },
                ],
                type: 'Export',
                locktime: 0,
            };
            const isVerified = await basecoin.verifyTransaction({ txParams, txPrebuild });
            isVerified.should.equal(true);
        });
        it('should verify an export from P-chain transaction', async () => {
            const txPrebuild = {
                txHex: exportInP_1.EXPORT_IN_P.fullSigntxHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: '',
                        amount: exportInP_1.EXPORT_IN_P.amount,
                    },
                ],
                type: 'Export',
                locktime: 0,
            };
            const isVerified = await basecoin.verifyTransaction({ txParams, txPrebuild });
            isVerified.should.equal(true);
        });
        it('should verify an import to C-chain transaction', async () => {
            const txPrebuild = {
                txHex: importInC_1.IMPORT_IN_C.fullSigntxHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: importInC_1.IMPORT_IN_C.to,
                        amount: '1',
                    },
                ],
                type: 'ImportToC',
                locktime: 0,
            };
            const isVerified = await basecoin.verifyTransaction({ txParams, txPrebuild });
            isVerified.should.equal(true);
        });
        it('should verify an import to P-chain transaction', async () => {
            const txPrebuild = {
                txHex: importInP_1.IMPORT_IN_P.signedHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [],
                type: 'Import',
                locktime: 0,
            };
            const isVerified = await basecoin.verifyTransaction({ txParams, txPrebuild });
            isVerified.should.equal(true);
        });
        it('should fail to verify export transaction with wrong amount', async () => {
            const txPrebuild = {
                txHex: exportInC_1.EXPORT_IN_C.signedHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: '',
                        amount: '999999999999',
                    },
                ],
                type: 'Export',
                locktime: 0,
            };
            await basecoin
                .verifyTransaction({ txParams, txPrebuild })
                .should.be.rejectedWith(/Tx total amount .* does not match with expected total amount/);
        });
        it('should fail to verify transaction with wrong type', async () => {
            const txPrebuild = {
                txHex: exportInC_1.EXPORT_IN_C.signedHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [
                    {
                        address: '',
                        amount: exportInC_1.EXPORT_IN_C.amount,
                    },
                ],
                type: 'Import',
                locktime: 0,
            };
            await basecoin
                .verifyTransaction({ txParams, txPrebuild })
                .should.be.rejectedWith('Tx type does not match with expected txParams type');
        });
        it('should fail to verify transaction without txHex', async () => {
            const txPrebuild = {
                txInfo: {},
            };
            const txParams = {
                recipients: [],
                type: 'Export',
                locktime: 0,
            };
            await basecoin
                .verifyTransaction({ txParams, txPrebuild })
                .should.be.rejectedWith('missing required tx prebuild property txHex');
        });
        it('should fail to verify transaction with invalid txHex', async () => {
            const txPrebuild = {
                txHex: 'invalidhex',
                txInfo: {},
            };
            const txParams = {
                recipients: [],
                type: 'Export',
                locktime: 0,
            };
            await basecoin
                .verifyTransaction({ txParams, txPrebuild })
                .should.be.rejectedWith('Invalid transaction: Raw transaction is not hex string');
        });
        it('should fail to verify import to C-chain without recipients', async () => {
            const txPrebuild = {
                txHex: importInC_1.IMPORT_IN_C.fullSigntxHex,
                txInfo: {},
            };
            const txParams = {
                recipients: [],
                type: 'ImportToC',
                locktime: 0,
            };
            await basecoin
                .verifyTransaction({ txParams, txPrebuild })
                .should.be.rejectedWith('Expected 1 recipient in import transaction');
        });
    });
    describe('Address Validation', () => {
        it('should validate mainnet P-chain address', function () {
            basecoin.isValidAddress(account_1.SEED_ACCOUNT.addressMainnet).should.be.true();
        });
        it('should validate testnet P-chain address', function () {
            basecoin.isValidAddress(account_1.SEED_ACCOUNT.addressTestnet).should.be.true();
        });
        it('should validate array of P-chain addresses', function () {
            basecoin.isValidAddress(exportInC_1.EXPORT_IN_C.pAddresses).should.be.true();
        });
        it('should validate tilde-separated multisig address', function () {
            const multiSigAddress = exportInC_1.EXPORT_IN_C.pAddresses.join('~');
            basecoin.isValidAddress(multiSigAddress).should.be.true();
        });
        it('should validate C-chain hex address', function () {
            basecoin.isValidAddress(exportInC_1.EXPORT_IN_C.cHexAddress).should.be.true();
        });
        it('should validate lowercase C-chain address', function () {
            basecoin.isValidAddress(importInC_1.IMPORT_IN_C.to.toLowerCase()).should.be.true();
        });
        it('should fail to validate undefined address', function () {
            basecoin.isValidAddress(undefined).should.be.false();
        });
        it('should fail to validate empty string', function () {
            basecoin.isValidAddress('').should.be.false();
        });
        it('should fail to validate invalid address', function () {
            basecoin.isValidAddress('invalid-address').should.be.false();
        });
        it('should fail to validate array with invalid address', function () {
            const addresses = [...exportInC_1.EXPORT_IN_C.pAddresses, 'invalid'];
            basecoin.isValidAddress(addresses).should.be.false();
        });
    });
    describe('Wallet Address Verification', () => {
        it('should verify wallet address with matching keychains', async () => {
            const keyPairs = [{ pub: account_1.SEED_ACCOUNT.publicKey }, { pub: account_1.ACCOUNT_1.publicKey }, { pub: account_1.ACCOUNT_2.publicKey }];
            // Derive addresses from public keys to ensure they match
            const derivedAddresses = keyPairs.map((kp) => new FlrpLib.KeyPair({ pub: kp.pub }).getAddress('testnet'));
            const address = derivedAddresses.join('~');
            const isValid = await basecoin.isWalletAddress({
                address,
                keychains: keyPairs,
            });
            isValid.should.be.true();
        });
        it('should throw for address with wrong number of keychains', async () => {
            const address = account_1.SEED_ACCOUNT.addressTestnet;
            await assert_1.default.rejects(async () => basecoin.isWalletAddress({
                address,
                keychains: [{ pub: account_1.SEED_ACCOUNT.publicKey }],
            }), /Invalid keychains/);
        });
        it('should throw for invalid address', async () => {
            await assert_1.default.rejects(async () => basecoin.isWalletAddress({
                address: 'invalid',
                keychains,
            }), /invalid address/);
        });
        it('should throw when address length does not match keychain length', async () => {
            const address = account_1.SEED_ACCOUNT.addressTestnet;
            await assert_1.default.rejects(async () => basecoin.isWalletAddress({
                address,
                keychains,
            }));
        });
        it('should throw when addresses do not match keychains', async () => {
            // Use addresses that don't match the keychains
            const address = exportInC_1.EXPORT_IN_C.pAddresses.join('~');
            await assert_1.default.rejects(async () => basecoin.isWalletAddress({
                address,
                keychains,
            }));
        });
    });
    describe('Recovery Signature', () => {
        it('should recover signature from signed message', async () => {
            const message = Buffer.from(account_1.SEED_ACCOUNT.message, 'utf8');
            const privateKey = Buffer.from(account_1.SEED_ACCOUNT.privateKey, 'hex');
            const signature = FlrpLib.Utils.createSignature(basecoin._staticsCoin.network, message, privateKey);
            const messageHash = FlrpLib.Utils.sha256(message);
            const recoveredPubKey = basecoin.recoverySignature(messageHash, signature);
            recoveredPubKey.should.be.instanceOf(Buffer);
            recoveredPubKey.length.should.equal(33);
        });
    });
    describe('Delegation Transaction Validation', () => {
        describe('validateDelegationTx', () => {
            const validStakingOptions = {
                nodeID: 'NodeID-AK7sPBsZM9rQwse23aLhEEBPHZD5gkLrL',
                amount: '50000000000000', // 50,000 FLR in nFLR
                durationSeconds: 86400 * 14, // 14 days
                rewardAddress: 'C-costwo1uyp5n76gjqltrddur7qlrsmt3kyh8fnrmwhqk7',
            };
            const validExplainedTx = {
                outputAmount: '50000000000000',
                type: sdk_core_1.TransactionType.AddPermissionlessDelegator,
                inputs: [],
                outputs: [],
                changeOutputs: [],
                changeAmount: '0',
                fee: { fee: '1000' },
            };
            it('should validate a correct delegation transaction', () => {
                const txParams = { stakingOptions: validStakingOptions };
                assert_1.default.doesNotThrow(() => basecoin.validateDelegationTx(txParams, validExplainedTx));
            });
            it('should throw when stakingOptions is missing', () => {
                const txParams = {};
                assert_1.default.throws(() => basecoin.validateDelegationTx(txParams, validExplainedTx), /Delegation transaction requires stakingOptions/);
            });
            it('should throw when nodeID has invalid format', () => {
                const txParams = {
                    stakingOptions: {
                        ...validStakingOptions,
                        nodeID: 'InvalidPrefix-123',
                    },
                };
                assert_1.default.throws(() => basecoin.validateDelegationTx(txParams, validExplainedTx), /Invalid nodeID format/);
            });
            it('should throw when amount is missing', () => {
                const txParams = {
                    stakingOptions: {
                        ...validStakingOptions,
                        amount: undefined,
                    },
                };
                assert_1.default.throws(() => basecoin.validateDelegationTx(txParams, validExplainedTx), /Delegation transaction requires amount/);
            });
            it('should throw when durationSeconds is missing', () => {
                const txParams = {
                    stakingOptions: {
                        ...validStakingOptions,
                        durationSeconds: undefined,
                    },
                };
                assert_1.default.throws(() => basecoin.validateDelegationTx(txParams, validExplainedTx), /Delegation transaction requires durationSeconds/);
            });
            it('should throw when rewardAddress is missing', () => {
                const txParams = {
                    stakingOptions: {
                        ...validStakingOptions,
                        rewardAddress: undefined,
                    },
                };
                assert_1.default.throws(() => basecoin.validateDelegationTx(txParams, validExplainedTx), /Delegation transaction requires rewardAddress/);
            });
            it('should throw when outputAmount does not match expected amount', () => {
                const txParams = { stakingOptions: validStakingOptions };
                const mismatchedExplainedTx = {
                    ...validExplainedTx,
                    outputAmount: '60000000000000', // Different from expected 50000000000000
                };
                assert_1.default.throws(() => basecoin.validateDelegationTx(txParams, mismatchedExplainedTx), /Delegation amount mismatch/);
            });
        });
    });
});
