// @bitgo/sdk-coin-eth/dist/test/unit/transactionBuilder/flushTokens.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../../src");
const getBuilder_1 = require("../getBuilder");
// eslint-disable-next-line import/no-internal-modules
const transactionBuilder_1 = require("@bitgo/abstract-eth/test/unit/transactionBuilder");
// Run the shared flush NFT tests from abstract-eth
describe('ETH Flush NFT Tests (from abstract-eth)', () => {
    (0, transactionBuilder_1.runFlushNftTests)('eth', getBuilder_1.getBuilder);
});
describe('Eth Transaction builder flush tokens (ETH-specific)', function () {
    const defaultKeyPair = new src_1.KeyPair({
        prv: 'FAC4D04AA0025ECF200D74BC9B5E4616E4B8338B69B61362AAAD49F76E68EF28',
    });
    const buildTransaction = async function (details) {
        const txBuilder = (0, getBuilder_1.getBuilder)('teth');
        txBuilder.type(sdk_core_1.TransactionType.FlushTokens);
        if (details.forwarderAddress !== undefined) {
            txBuilder.forwarderAddress(details.forwarderAddress);
        }
        if (details.tokenAddress !== undefined) {
            txBuilder.tokenAddress(details.tokenAddress);
        }
        if (details.fee !== undefined) {
            txBuilder.fee(details.fee);
        }
        if (details.contractAddress !== undefined) {
            txBuilder.contract(details.contractAddress);
        }
        if (details.counter !== undefined) {
            txBuilder.counter(details.counter);
        }
        if (details.key !== undefined) {
            txBuilder.sign({ key: details.key.getKeys().prv });
        }
        if (details.forwarderVersion !== undefined) {
            txBuilder.forwarderVersion(details.forwarderVersion);
        }
        return await txBuilder.build();
    };
    describe('should build', () => {
        it('a wallet flush forwarder transaction', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 1,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            });
            tx.type.should.equal(sdk_core_1.TransactionType.FlushTokens);
            const txJson = tx.toJson();
            txJson.gasLimit.should.equal('1000');
            txJson._type.should.equals(src_1.ETHTransactionType.LEGACY);
            txJson.gasPrice.should.equal('10');
            should_1.default.equal(txJson.nonce, 1);
            txJson.data.should.startWith(src_1.flushForwarderTokensMethodId);
        });
        it('a wallet flush forwarder transaction with eip1559 fee model', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '30',
                    eip1559: {
                        maxPriorityFeePerGas: '5',
                        maxFeePerGas: '30',
                    },
                    gasLimit: '1000',
                },
                counter: 1,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            });
            tx.type.should.equal(sdk_core_1.TransactionType.FlushTokens);
            const txJson = tx.toJson();
            txJson.gasLimit.should.equal('1000');
            txJson._type.should.equals(src_1.ETHTransactionType.EIP1559);
            txJson.maxFeePerGas.should.equal('30');
            txJson.maxPriorityFeePerGas.should.equal('5');
            should_1.default.equal(txJson.nonce, 1);
            txJson.data.should.startWith(src_1.flushForwarderTokensMethodId);
        });
        it('a wallet flush forwarder transaction with nonce 0', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            });
            tx.type.should.equal(sdk_core_1.TransactionType.FlushTokens);
            const txJson = tx.toJson();
            txJson.gasLimit.should.equal('1000');
            txJson._type.should.equals(src_1.ETHTransactionType.LEGACY);
            txJson.gasPrice.should.equal('10');
            should_1.default.equal(txJson.nonce, 0);
            txJson.data.should.startWith(src_1.flushForwarderTokensMethodId);
        });
        it('a wallet flush forwarder transaction with forwarder Version 4', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                forwarderVersion: 4,
            });
            tx.type.should.equal(sdk_core_1.TransactionType.FlushTokens);
            const txJson = tx.toJson();
            txJson.gasLimit.should.equal('1000');
            txJson._type.should.equals(src_1.ETHTransactionType.LEGACY);
            should_1.default.equal(txJson.nonce, 0);
            txJson.data.should.startWith(src_1.flushForwarderTokensMethodIdV4);
        });
        it('decode wallet flush forwarder transaction with forwarder Version 4', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                forwarderVersion: 4,
            });
            const txBuiderFromRaw = (0, getBuilder_1.getBuilder)('teth');
            txBuiderFromRaw.from(tx.toBroadcastFormat());
            const result = await txBuiderFromRaw.build();
            const txJson = result.toJson();
            txJson.to.should.equal('0x53b8e91bb3b8f618b5f01004ef108f134f219573');
            txJson.data.should.containEql('bcf935d206ca32929e1b887a07ed240f0d8ccd22');
        });
        it('decode wallet flush forwarder transaction with forwarder Version < 4', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
                forwarderVersion: 2,
            });
            const txBuiderFromRaw = (0, getBuilder_1.getBuilder)('teth');
            txBuiderFromRaw.from(tx.toBroadcastFormat());
            const result = await txBuiderFromRaw.build();
            const txJson = result.toJson();
            txJson.to.should.equal('0x8f977e912ef500548a0c3be6ddde9899f1199b81');
            txJson.data.should.containEql('bcf935d206ca32929e1b887a07ed240f0d8ccd22');
            txJson.data.should.containEql('53b8e91bb3b8f618b5f01004ef108f134f219573');
        });
        it('an unsigned flush transaction from serialized', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            });
            const serialized = tx.toBroadcastFormat();
            // now rebuild from the signed serialized tx and make sure it stays the same
            const newTxBuilder = (0, getBuilder_1.getBuilder)('teth');
            newTxBuilder.from(serialized);
            const newTx = await newTxBuilder.build();
            should_1.default.equal(newTx.toBroadcastFormat(), serialized);
            newTx.toJson().data.should.startWith(src_1.flushForwarderTokensMethodId);
            should_1.default.equal(newTx.toJson().v, '0x77');
        });
        it('a signed flush transaction from serialized', async () => {
            const tx = await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
                key: defaultKeyPair,
            });
            const serialized = tx.toBroadcastFormat();
            // now rebuild from the signed serialized tx and make sure it stays the same
            const newTxBuilder = (0, getBuilder_1.getBuilder)('teth');
            newTxBuilder.from(serialized);
            const newTx = await newTxBuilder.build();
            should_1.default.equal(newTx.toBroadcastFormat(), serialized);
            should_1.default.equal(newTx.id, '0x5bfd4c6b9ae9dfbc9e062247c0debe4e1c6870a76c420ce7f6e335390420dcd6');
            const txJson = newTx.toJson();
            should_1.default.exist(txJson.v);
            should_1.default.exist(txJson.r);
            should_1.default.exist(txJson.s);
            should_1.default.exist(txJson.from);
        });
    });
    describe('should fail to build', () => {
        it('a transaction without fee', async () => {
            await buildTransaction({
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            }).should.be.rejectedWith('Invalid transaction: missing fee');
        });
        it('a transaction without contractAddress', async () => {
            await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
            }).should.be.rejectedWith('Invalid transaction: missing contract address');
        });
        it('a wallet flush forwarder V4 transaction with different contract address', async () => {
            await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
                forwarderVersion: 4,
            }).should.be.rejectedWith('Invalid contract address: 0x8f977e912ef500548a0c3be6ddde9899f1199b81');
        });
        it('a transaction without tokenAddress', async () => {
            await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                counter: 0,
                forwarderAddress: '0x53b8e91bb3b8f618b5f01004ef108f134f219573',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            }).should.be.rejectedWith('Invalid transaction: missing token address');
        });
        it('a transaction without forwarderAddress', async () => {
            await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                counter: 0,
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            }).should.be.rejectedWith('Invalid transaction: missing forwarder address');
        });
        it('a transaction with invalid counter', async () => {
            await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                counter: -1,
                tokenAddress: '0xbcf935d206ca32929e1b887a07ed240f0d8ccd22',
                contractAddress: '0x8f977e912ef500548a0c3be6ddde9899f1199b81',
            }).should.be.rejectedWith('Invalid counter: -1');
        });
    });
});
