// @bitgo/sdk-coin-eth/dist/test/unit/transactionBuilder/walletInitialization.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const should_1 = __importDefault(require("should"));
const sdk_core_1 = require("@bitgo/sdk-core");
const src_1 = require("../../../src");
const testData = __importStar(require("../../resources/eth"));
const getBuilder_1 = require("../getBuilder");
describe('Eth Transaction builder wallet initialization', function () {
    const sourcePrv = 'xprv9s21ZrQH143K3D8TXfvAJgHVfTEeQNW5Ys9wZtnUZkqPzFzSjbEJrWC1vZ4GnXCvR7rQL2UFX3RSuYeU9MrERm1XBvACow7c36vnz5iYyj2';
    const pub1 = 'xpub661MyMwAqRbcGpyL5QvWah4XZYHuTK21mSQ4NVwYaX67A35Kzb42nmTdf2WArW4tettXrWpfpwFbEFdEVqcSvnHLB8F6p1D41ssmbnRMXpc';
    const pub2 = 'xpub661MyMwAqRbcFWzoz8qnYRDYEFQpPLYwxVFoG6WLy3ck5ZupRGJTG4ju6yGb7Dj3ey6GsC4kstLRER2nKzgjLtmxyPgC4zHy7kVhUt6yfGn';
    const defaultKeyPair = new src_1.KeyPair({
        prv: 'FAC4D04AA0025ECF200D74BC9B5E4616E4B8338B69B61362AAAD49F76E68EF28',
    });
    const buildTransaction = async function (details) {
        const txBuilder = (0, getBuilder_1.getBuilder)('teth');
        if (details.type !== undefined) {
            txBuilder.type(details.type);
        }
        if (details.fee !== undefined) {
            txBuilder.fee(details.fee);
        }
        if (details.counter !== undefined) {
            txBuilder.counter(details.counter);
        }
        if (details.owners !== undefined) {
            for (const owner of details.owners) {
                txBuilder.owner(owner);
            }
        }
        return await txBuilder.build();
    };
    describe('should build', () => {
        it('a wallet initialization transaction', async () => {
            const tx = await buildTransaction({
                type: sdk_core_1.TransactionType.WalletInitialization,
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                owners: [
                    new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                    new src_1.KeyPair({ pub: pub1 }).getAddress(),
                    new src_1.KeyPair({ pub: pub2 }).getAddress(),
                ],
                counter: 1,
            });
            tx.type.should.equal(sdk_core_1.TransactionType.WalletInitialization);
            const txJson = tx.toJson();
            txJson.gasLimit.should.equal('1000');
            txJson._type.should.equals(src_1.ETHTransactionType.LEGACY);
            txJson.gasPrice.should.equal('10');
            should_1.default.equal(txJson.nonce, 1);
            should_1.default.equal(txJson.chainId, 42);
            should_1.default.equal(txJson.v, '0x77');
        });
        it('a wallet initialization transaction with nonce 0', async () => {
            const tx = await buildTransaction({
                type: sdk_core_1.TransactionType.WalletInitialization,
                fee: {
                    fee: '10',
                    gasLimit: '1000',
                },
                owners: [
                    new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                    new src_1.KeyPair({ pub: pub1 }).getAddress(),
                    new src_1.KeyPair({ pub: pub2 }).getAddress(),
                ],
                counter: 0,
            });
            tx.type.should.equal(sdk_core_1.TransactionType.WalletInitialization);
            const txJson = tx.toJson();
            txJson.gasLimit.should.equal('1000');
            txJson._type.should.equals(src_1.ETHTransactionType.LEGACY);
            txJson.gasPrice.should.equal('10');
            should_1.default.equal(txJson.nonce, 0);
            should_1.default.equal(txJson.chainId, 42);
            should_1.default.equal(txJson.v, '0x77');
        });
        it('an unsigned init transaction from serialized with 0-prefixed address', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.counter(1);
            txBuilder.fee({
                fee: '10000000000',
                gasLimit: '2000000',
            });
            txBuilder.owner('0x6461EC4E9dB87CFE2aeEc7d9b02Aa264edFbf41f');
            txBuilder.owner('0xf10C8f42BD63D0AeD3338A6B2b661BC6D9fa7C44');
            txBuilder.owner('0x07ee8b845b8bf0e807e096d6b1599b121b82cbe1');
            const tx = await txBuilder.build();
            const serialized = tx.toBroadcastFormat();
            // now rebuild from the signed serialized tx and make sure it stays the same
            const newTxBuilder = (0, getBuilder_1.getBuilder)('eth');
            newTxBuilder.from(serialized);
            const newTx = await newTxBuilder.build();
            should_1.default.equal(newTx.toBroadcastFormat(), serialized);
        });
        it('an unsigned init transaction from serialized', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.counter(1);
            txBuilder.fee({
                fee: '10000000000',
                gasLimit: '2000000',
            });
            txBuilder.owner('0x6461EC4E9dB87CFE2aeEc7d9b02Aa264edFbf41f');
            txBuilder.owner('0xf10C8f42BD63D0AeD3338A6B2b661BC6D9fa7C44');
            txBuilder.owner('0xa4b5666FB4fFEA84Dd848845E1114b84146de4b3');
            const tx = await txBuilder.build();
            const serialized = tx.toBroadcastFormat();
            // now rebuild from the signed serialized tx and make sure it stays the same
            const newTxBuilder = (0, getBuilder_1.getBuilder)('eth');
            newTxBuilder.from(serialized);
            const newTx = await newTxBuilder.build();
            should_1.default.equal(newTx.toBroadcastFormat(), serialized);
        });
        it('a signed init transaction from serialized', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('teth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.counter(1);
            txBuilder.fee({
                fee: '10000000000',
                gasLimit: '2000000',
            });
            txBuilder.owner('0x6461EC4E9dB87CFE2aeEc7d9b02Aa264edFbf41f');
            txBuilder.owner('0xf10C8f42BD63D0AeD3338A6B2b661BC6D9fa7C44');
            txBuilder.owner('0xa4b5666FB4fFEA84Dd848845E1114b84146de4b3');
            txBuilder.sign({ key: defaultKeyPair.getKeys().prv });
            const tx = await txBuilder.build();
            const serialized = tx.toBroadcastFormat();
            // now rebuild from the signed serialized tx and make sure it stays the same
            const newTxBuilder = (0, getBuilder_1.getBuilder)('teth');
            newTxBuilder.from(serialized);
            const newTx = await newTxBuilder.build();
            should_1.default.equal(newTx.toBroadcastFormat(), serialized);
            should_1.default.equal(newTx.id, '0xef04334f21ba844100b9aee6485f14afe177e88308ad2b8c1b0878576a31d47b');
            const txJson = newTx.toJson();
            should_1.default.exist(txJson.v);
            should_1.default.exist(txJson.r);
            should_1.default.exist(txJson.s);
            should_1.default.exist(txJson.from);
        });
        it('correct transaction id', async () => {
            const newTxBuilder = (0, getBuilder_1.getBuilder)('eth');
            newTxBuilder.from(testData.WALLET_INITIALIZATION);
            const newTx = await newTxBuilder.build();
            should_1.default.equal(newTx.toJson().id, '0xc65f9802df3b559b297779ec06d3e71ba7f5b1b47cc961ad2efba54d82347bec');
        });
    });
    describe('should fail to build', () => {
        it('an unsupported type of transaction', async () => {
            await buildTransaction({
                type: sdk_core_1.TransactionType.AccountUpdate,
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                counter: 0,
            }).should.be.rejectedWith('Unsupported transaction type');
        });
        it('a transaction without fee', async () => {
            await buildTransaction({
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                counter: 0,
            }).should.be.rejectedWith('Invalid transaction: missing fee');
        });
        it('a wallet initialization the wrong number of owners', async () => {
            await buildTransaction({
                type: sdk_core_1.TransactionType.WalletInitialization,
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                owners: [new src_1.KeyPair({ pub: pub1 }).getAddress(), new src_1.KeyPair({ pub: pub2 }).getAddress()],
                counter: 0,
            }).should.be.rejectedWith('Invalid transaction: wrong number of owners -- required: 3, found: 2');
            await buildTransaction({
                type: sdk_core_1.TransactionType.WalletInitialization,
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                owners: [
                    new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                    new src_1.KeyPair({ pub: pub1 }).getAddress(),
                    new src_1.KeyPair({ pub: pub1 }).getAddress(),
                    new src_1.KeyPair({ pub: pub2 }).getAddress(),
                ],
                counter: 0,
            }).should.be.rejectedWith('Repeated owner address: ' + new src_1.KeyPair({ pub: pub1 }).getAddress());
            await buildTransaction({
                type: sdk_core_1.TransactionType.WalletInitialization,
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                owners: [],
                counter: 0,
            }).should.be.rejectedWith('Invalid transaction: wrong number of owners -- required: 3, found: 0');
        });
        it('a transaction with invalid counter', async () => {
            await buildTransaction({
                fee: {
                    fee: '10',
                    gasLimit: '10',
                },
                source: new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                counter: -1,
                owners: [
                    new src_1.KeyPair({ prv: sourcePrv }).getAddress(),
                    new src_1.KeyPair({ pub: pub1 }).getAddress(),
                    new src_1.KeyPair({ pub: pub2 }).getAddress(),
                ],
            }).should.be.rejectedWith('Invalid counter: -1');
        });
    });
    describe('should fail to sign', () => {
        it('a wallet initialization transaction without owners', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.fee({
                fee: '10',
                gasLimit: '1000',
            });
            txBuilder.counter(1);
            assert_1.default.throws(() => txBuilder.sign({ key: defaultKeyPair.getKeys().prv }));
        });
        it('a signed wallet initialization transaction', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.fee({
                fee: '10',
                gasLimit: '1000',
            });
            txBuilder.counter(1);
            txBuilder.owner(new src_1.KeyPair({ pub: pub1 }).getAddress());
            txBuilder.owner(new src_1.KeyPair({ pub: pub2 }).getAddress());
            txBuilder.owner(new src_1.KeyPair({ prv: sourcePrv }).getAddress());
            txBuilder.sign({ key: defaultKeyPair.getKeys().prv });
            assert_1.default.throws(() => txBuilder.sign({ key: defaultKeyPair.getKeys().prv }), new RegExp('Cannot sign multiple times a non send-type transaction'));
        });
    });
    describe('should validate', () => {
        it('an address', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.validateAddress(testData.VALID_ADDRESS);
            assert_1.default.throws(() => txBuilder.validateAddress(testData.INVALID_ADDRESS), new RegExp('Invalid address ' + testData.INVALID_ADDRESS.address));
        });
        it('value should be greater than zero', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            assert_1.default.throws(() => txBuilder.fee({ fee: '-10' }));
            should_1.default.doesNotThrow(() => txBuilder.fee({ fee: '10' }));
        });
        it('a private key', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            assert_1.default.throws(() => txBuilder.validateKey({ key: 'abc' }), /Invalid key/);
            assert_1.default.throws(() => txBuilder.validateKey({ key: testData.PUBLIC_KEY }), /Invalid key/);
            should_1.default.doesNotThrow(() => txBuilder.validateKey({ key: testData.PRIVATE_KEY }));
        });
        it('a raw transaction', async () => {
            const builder = (0, getBuilder_1.getBuilder)('eth');
            should_1.default.doesNotThrow(() => builder.from(testData.TX_BROADCAST));
            should_1.default.doesNotThrow(() => builder.from(testData.TX_JSON));
            assert_1.default.throws(() => builder.from('0x00001000'), /There was error in decoding the hex string/);
            assert_1.default.throws(() => builder.from(''), /Raw transaction is empty/);
            assert_1.default.throws(() => builder.from('pqrs'), /There was error in parsing the JSON string/);
            assert_1.default.throws(() => builder.from(1234), /Transaction is not a hex string or stringified json/);
        });
        it('a transaction to build', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.counter(undefined);
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            assert_1.default.throws(() => txBuilder.validateTransaction(), /Invalid transaction: missing fee/);
            txBuilder.fee({
                fee: '10',
                gasLimit: '1000',
            });
            assert_1.default.throws(() => txBuilder.validateTransaction(), /Invalid transaction: missing address counter/);
            assert_1.default.throws(() => txBuilder.validateTransaction(), /Invalid transaction: missing address counter/);
            const source = {
                prv: sourcePrv,
            };
            const sourceKeyPair = new src_1.KeyPair(source);
            assert_1.default.throws(() => txBuilder.validateTransaction(), /Invalid transaction: missing address counter/);
            txBuilder.counter(1);
            assert_1.default.throws(() => txBuilder.validateTransaction(), /wrong number of owners -- required: 3, found: 0/);
            txBuilder.owner(sourceKeyPair.getAddress());
            assert_1.default.throws(() => txBuilder.validateTransaction(), /wrong number of owners -- required: 3, found: 1/);
            txBuilder.owner(new src_1.KeyPair({ pub: pub1 }).getAddress());
            assert_1.default.throws(() => txBuilder.validateTransaction(), /wrong number of owners -- required: 3, found: 2/);
            txBuilder.owner(new src_1.KeyPair({ pub: pub2 }).getAddress());
            should_1.default.doesNotThrow(() => txBuilder.validateTransaction());
        });
    });
    describe('set owner', () => {
        it('should be wallet initializaion', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.Send);
            const sourceKeyPair = new src_1.KeyPair({ prv: sourcePrv });
            assert_1.default.throws(() => txBuilder.owner(sourceKeyPair.getAddress()), new RegExp('Multisig wallet owner can only be set for initialization transactions'));
        });
        it('should be only 3 owners', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.fee({
                fee: '10',
                gasLimit: '1000',
            });
            const sourceKeyPair = new src_1.KeyPair({ prv: sourcePrv });
            txBuilder.counter(1);
            txBuilder.owner(sourceKeyPair.getAddress());
            txBuilder.owner('0x7325A3F7d4f9E86AE62Cf742426078C3755730d5');
            txBuilder.owner('0x603e077acd3F01e81b95fB92ce42FF60dFf3D4C7');
            assert_1.default.throws(() => txBuilder.owner('0x1A88Ee4Bc80BE080fC91AC472Af2F59260695060'), new RegExp('A maximum of 3 owners can be set for a multisig wallet'));
        });
        it('should be a valid address', () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('eth');
            txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
            txBuilder.fee({
                fee: '10',
                gasLimit: '1000',
            });
            txBuilder.counter(1);
            assert_1.default.throws(() => txBuilder.owner('0x7325A3F7d4f9E86AE62C'), /Invalid address/);
        });
    });
    it('Should getting same tx hash from raw tx', async function () {
        const txBuilder = (0, getBuilder_1.getBuilder)('eth');
        txBuilder.fee({
            fee: '280000000000',
            gasLimit: '7000000',
        });
        txBuilder.counter(1);
        txBuilder.type(sdk_core_1.TransactionType.WalletInitialization);
        txBuilder.owner(new src_1.KeyPair({ pub: pub1 }).getAddress());
        txBuilder.owner(new src_1.KeyPair({ pub: pub2 }).getAddress());
        txBuilder.owner(new src_1.KeyPair({ prv: sourcePrv }).getAddress());
        txBuilder.sign({ key: defaultKeyPair.getKeys().prv });
        const tx = await txBuilder.build();
        const txBuiderFromRaw = (0, getBuilder_1.getBuilder)('eth');
        txBuiderFromRaw.from(tx.toBroadcastFormat());
        const txFromRaw = await txBuiderFromRaw.build();
        should_1.default.deepEqual(tx.id, txFromRaw.id);
    });
    describe('Recovery Wallet', function () {
        it('should build and sign using txHex', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('gteth');
            const txHex = testData.RECOVERY_WALLET_DEPLOYMENT_TXHEX;
            txBuilder.from(txHex);
            const xpriv = 'xprv9s21ZrQH143K2YKSzZa2fv7QTzZ5Ui45ZUzygDwhfeHodwHbWtUUSKqikMKZR9Z751NMekrjXKBykb8mhvoTrKMRodKBvNDH4XKPkccF1K4';
            txBuilder.sign({ key: xpriv });
            const signedTx = await txBuilder.build();
            should_1.default.exists(signedTx.toBroadcastFormat());
        });
        it('succeed when build via contract call', async () => {
            const txBuilder = (0, getBuilder_1.getBuilder)('hteth');
            txBuilder.type(sdk_core_1.TransactionType.ContractCall);
            txBuilder.fee({
                eip1559: {
                    maxFeePerGas: '100',
                    maxPriorityFeePerGas: '10',
                },
                fee: '100',
                gasLimit: '10000',
            });
            const data = testData.SEND_FUNDS_METHOD_CALL;
            txBuilder.counter(1);
            txBuilder.data(data);
            txBuilder.contract('0xd536f4b9f9127a39f19820ca18baac7cd157471f');
            const tx = await txBuilder.build();
            const txhex = tx.toBroadcastFormat();
            const builderFrom = (0, getBuilder_1.getBuilder)('hteth');
            builderFrom.from(txhex);
            const txFrom = await builderFrom.build();
            should_1.default.deepEqual(tx.id, txFrom.id);
            should_1.default.deepEqual(tx.toJson(), txFrom.toJson());
        });
    });
});
