// @bitgo/sdk-coin-apt/dist/src/lib/transactionBuilder/customTransactionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const transactionBuilder_1 = require("./transactionBuilder");
const customTransaction_1 = require("../transaction/customTransaction");
const validation_1 = require("../utils/validation");
/**
 * Builder for Aptos custom transactions.
 * Allows building transactions with any entry function call.
 */
class CustomTransactionBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new customTransaction_1.CustomTransaction(_coinConfig);
    }
    /**
     * @returns {TransactionType} The transaction type
     */
    get transactionType() {
        return sdk_core_1.TransactionType.CustomTx;
    }
    /**
     * Set the custom transaction parameters
     *
     * @param {CustomTransactionParams} params - Custom transaction parameters
     * @returns {CustomTransactionBuilder} The builder instance
     */
    customTransaction(params) {
        if (!params.moduleName) {
            throw new sdk_core_1.BuildTransactionError('Missing module name');
        }
        if (!params.functionName) {
            throw new sdk_core_1.BuildTransactionError('Missing function name');
        }
        this._transaction.setCustomTransactionParams(params);
        return this;
    }
    /**
     * Set the entry function ABI
     *
     * @param {EntryFunctionABI} abi - The ABI definition for the entry function
     * @returns {CustomTransactionBuilder} The builder instance
     */
    entryFunctionAbi(abi) {
        this._transaction.setEntryFunctionAbi(abi);
        return this;
    }
    /**
     * @inheritdoc
     */
    assetId(assetId) {
        // Asset ID is optional for custom transactions
        this._transaction.assetId = assetId;
        return this;
    }
    /**
     * Check if a transaction payload is valid for this builder
     *
     * @param {TransactionPayload} payload - The transaction payload to check
     * @returns {boolean} True if the payload is valid for this builder
     */
    isValidTransactionPayload(payload) {
        // Basic payload type check
        if (!(payload instanceof ts_sdk_1.TransactionPayloadEntryFunction)) {
            return false;
        }
        try {
            const entryFunction = payload.entryFunction;
            // Validate module and function identifiers
            const moduleAddress = entryFunction.module_name.address.toString();
            const moduleIdentifier = entryFunction.module_name.name.identifier;
            const functionIdentifier = entryFunction.function_name.identifier;
            const moduleName = `${moduleAddress}::${moduleIdentifier}`;
            if (!(0, validation_1.isValidModuleName)(moduleName) || !(0, validation_1.isValidFunctionName)(functionIdentifier)) {
                return false;
            }
            return true;
        }
        catch (error) {
            return false;
        }
    }
    /**
     * Initialize the builder from an existing transaction
     *
     * @param {Transaction} tx - The transaction to initialize from
     */
    initBuilder(tx) {
        super.initBuilder(tx);
        if (tx instanceof customTransaction_1.CustomTransaction) {
            const params = tx.getCustomTransactionParams();
            // Only set custom transaction params if they are defined
            if (params.moduleName && params.functionName) {
                this.customTransaction(params);
            }
        }
    }
    /**
     * Build the transaction
     *
     * @returns {Promise<CustomTransaction>} The built transaction
     */
    async build() {
        const params = this._transaction.getCustomTransactionParams();
        if (!params.moduleName || !params.functionName) {
            throw new sdk_core_1.BuildTransactionError('Missing required module or function name');
        }
        return (await super.build());
    }
    /**
     * Get the transaction builder instance from an existing transaction
     *
     * @param {Transaction} tx - The transaction to convert
     * @returns {TransactionBuilder} The transaction builder
     */
    from(tx) {
        if (!(tx instanceof customTransaction_1.CustomTransaction)) {
            throw new sdk_core_1.BuildTransactionError('Invalid transaction type');
        }
        this.initBuilder(tx);
        return this;
    }
}
exports.CustomTransactionBuilder = CustomTransactionBuilder;
