// @bitgo/sdk-coin-apt/dist/src/lib/transactionBuilder/delegationPoolUnlockTransactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegationPoolUnlockTransactionBuilder = void 0;
const transactionBuilder_1 = require("./transactionBuilder");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const delegationPoolUnlockTransaction_1 = require("../transaction/delegationPoolUnlockTransaction");
class DelegationPoolUnlockTransactionBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.transaction = new delegationPoolUnlockTransaction_1.DelegationPoolUnlockTransaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingUnlock;
    }
    assetId(_assetId) {
        this.transaction.assetId = _assetId;
        return this;
    }
    validator(validatorAddress, amount) {
        this._transaction.validatorAddress = validatorAddress;
        this._transaction.amount = amount;
        return this;
    }
    isValidTransactionPayload(payload) {
        try {
            if (!this.isValidPayload(payload)) {
                return false;
            }
            const { entryFunction } = payload;
            const addressArg = entryFunction.args[0];
            const amountArg = entryFunction.args[1];
            return utils_1.default.fetchAndValidateRecipients(addressArg, amountArg).isValid;
        }
        catch (e) {
            return false;
        }
    }
    isValidPayload(payload) {
        return (payload instanceof ts_sdk_1.TransactionPayloadEntryFunction &&
            payload.entryFunction.args.length === 2 &&
            payload.entryFunction.type_args.length === 0);
    }
}
exports.DelegationPoolUnlockTransactionBuilder = DelegationPoolUnlockTransactionBuilder;
