// @bitgo/sdk-coin-apt/dist/src/lib/transactionBuilder/fungibleAssetTransferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FungibleAssetTransferBuilder = void 0;
const transactionBuilder_1 = require("./transactionBuilder");
const fungibleAssetTransfer_1 = require("../transaction/fungibleAssetTransfer");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = __importDefault(require("../utils"));
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const constants_1 = require("../constants");
const logger_1 = require("@bitgo/logger");
class FungibleAssetTransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this._transaction = new fungibleAssetTransfer_1.FungibleAssetTransfer(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.SendToken;
    }
    //TODO: Ticket: COIN-2941 : Check Statics based asset validation and if possible, implement it
    assetId(assetId) {
        this.validateAddress({ address: assetId });
        this.transaction.assetId = assetId;
        return this;
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('fungible asset transaction not defined');
        }
        super.validateTransaction(transaction);
        this.validateAddress({ address: transaction.assetId });
    }
    isValidTransactionPayload(payload) {
        try {
            if (!this.isValidPayload(payload)) {
                console.error('invalid transaction payload');
                return false;
            }
            const entryFunction = payload.entryFunction;
            const fungibleTokenAddress = entryFunction.args[0].toString();
            const addressArg = entryFunction.args[1];
            const amountArg = entryFunction.args[2];
            return (utils_1.default.isValidAddress(fungibleTokenAddress) && utils_1.default.fetchAndValidateRecipients(addressArg, amountArg).isValid);
        }
        catch (e) {
            logger_1.logger.error('invalid transaction payload', e);
            return false;
        }
    }
    isValidPayload(payload) {
        return (payload instanceof ts_sdk_1.TransactionPayloadEntryFunction &&
            payload.entryFunction.args.length === 3 &&
            (payload.entryFunction.type_args.length === 0 ||
                (payload.entryFunction.type_args.length === 1 &&
                    constants_1.FUNGIBLE_ASSET_TYPE_ARGUMENT === payload.entryFunction.type_args[0].toString())));
    }
}
exports.FungibleAssetTransferBuilder = FungibleAssetTransferBuilder;
