// @bitgo/sdk-coin-apt/dist/src/lib/transactionBuilder/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const transactionBuilder_1 = require("./transactionBuilder");
const sdk_core_1 = require("@bitgo/sdk-core");
const transferTransaction_1 = require("../transaction/transferTransaction");
const utils_1 = __importDefault(require("../utils"));
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const logger_1 = require("@bitgo/logger");
class TransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.transaction = new transferTransaction_1.TransferTransaction(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    validatePartsOfAssetId(assetId) {
        const parts = assetId.split('::');
        if (parts.length !== 3 || ('0x1' !== parts[0] && !utils_1.default.isValidAddress(parts[0]))) {
            throw new Error('Invalid asset ID');
        }
    }
    assetId(assetId) {
        this.validatePartsOfAssetId(assetId);
        this.transaction.assetId = assetId;
        return this;
    }
    isValidTransactionPayload(payload) {
        try {
            if (!this.isValidPayload(payload)) {
                console.error('invalid transaction payload');
                return false;
            }
            const entryFunction = payload.entryFunction;
            this.validatePartsOfAssetId(entryFunction.type_args[0].toString());
            const addressArg = entryFunction.args[0];
            const amountArg = entryFunction.args[1];
            return utils_1.default.fetchAndValidateRecipients(addressArg, amountArg).isValid;
        }
        catch (e) {
            logger_1.logger.error('invalid transaction payload', e);
            return false;
        }
    }
    isValidPayload(payload) {
        return (payload instanceof ts_sdk_1.TransactionPayloadEntryFunction &&
            payload.entryFunction.args.length === 2 &&
            payload.entryFunction.type_args.length === 1 &&
            payload.entryFunction.type_args[0].toString().length > 0);
    }
}
exports.TransferBuilder = TransferBuilder;
