// @bitgo/sdk-coin-apt/dist/src/lib/transactionBuilder/transactionBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const logger_1 = require("@bitgo/logger");
const utils_1 = __importDefault(require("../utils"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class TransactionBuilder extends sdk_core_1.BaseTransactionBuilder {
    constructor(coinConfig) {
        super(coinConfig);
    }
    getRebuiltTransaction() {
        return this._transaction;
    }
    getSequenceNumber() {
        return this.transaction.sequenceNumber;
    }
    /**
     * Initialize the transaction builder fields using the decoded transaction data
     *
     * @param {Transaction} tx the transaction data
     */
    initBuilder(tx) {
        this._transaction = tx;
    }
    /** @inheritdoc */
    get transaction() {
        return this._transaction;
    }
    /** @inheritdoc */
    set transaction(transaction) {
        this._transaction = transaction;
    }
    /**
     * Sets the sender of this transaction.
     *
     * @param {string} senderAddress the account that is sending this transaction
     * @returns {TransactionBuilder} This transaction builder
     */
    sender(senderAddress) {
        this.validateAddress({ address: senderAddress });
        this.transaction.sender = senderAddress;
        return this;
    }
    /**
     * @deprecated - use `recipients()`.
     */
    recipient(recipient) {
        this.validateAddress({ address: recipient.address });
        this.validateValue(new bignumber_js_1.default(recipient.amount));
        this.transaction.recipients = [recipient];
        return this;
    }
    recipients(recipients) {
        for (const recipient of recipients) {
            this.validateAddress({ address: recipient.address });
            this.validateValue(new bignumber_js_1.default(recipient.amount));
        }
        this.transaction.recipients = recipients;
        return this;
    }
    gasData(gasData) {
        this.validateGasData(gasData);
        this.transaction.maxGasAmount = gasData.maxGasAmount;
        this.transaction.gasUnitPrice = gasData.gasUnitPrice;
        this.transaction.gasUsed = gasData.gasUsed ?? 0;
        return this;
    }
    sequenceNumber(seqNo) {
        this.transaction.sequenceNumber = seqNo;
        return this;
    }
    expirationTime(expTimeSec) {
        this.transaction.expirationTime = expTimeSec;
        return this;
    }
    /** @inheritdoc */
    signImplementation(key) {
        throw new Error('Method not implemented.');
    }
    /** @inheritDoc */
    addSenderSignature(publicKey, signature) {
        this.transaction.addSenderSignature(publicKey, signature);
    }
    addFeePayerSignature(publicKey, signature) {
        this.transaction.addFeePayerSignature(publicKey, signature);
    }
    setIsSimulateTxn(value) {
        this.transaction.isSimulateTxn = value;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        this.transaction.fromRawTransaction(rawTransaction);
        this.transaction.transactionType = this.transactionType;
        return this.transaction;
    }
    /** @inheritdoc */
    async buildImplementation() {
        this.transaction.transactionType = this.transactionType;
        await this.transaction.build();
        return this.transaction;
    }
    // region Validators
    /** @inheritdoc */
    validateAddress(address, addressFormat) {
        if (!utils_1.default.isValidAddress(address.address)) {
            throw new sdk_core_1.BuildTransactionError('Invalid address ' + address.address);
        }
    }
    /** @inheritdoc */
    validateKey(key) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    validateTransaction(transaction) {
        if (!transaction) {
            throw new Error('transaction not defined');
        }
        this.validateAddress({ address: transaction.sender });
        for (const recipient of transaction.recipients) {
            this.validateAddress({ address: recipient.address });
            this.validateValue(new bignumber_js_1.default(recipient.amount));
        }
    }
    isValidRawTransaction(rawTransaction) {
        try {
            const signedTxn = utils_1.default.deserializeSignedTransaction(rawTransaction);
            const rawTxn = signedTxn.raw_txn;
            const senderAddress = rawTxn.sender.toString();
            return utils_1.default.isValidAddress(senderAddress) && this.isValidTransactionPayload(rawTxn.payload);
        }
        catch (e) {
            logger_1.logger.error('invalid raw transaction', e);
            return false;
        }
    }
    /** @inheritdoc */
    validateRawTransaction(rawTransaction) {
        if (!rawTransaction) {
            throw new sdk_core_1.ParseTransactionError('Invalid raw transaction: Undefined');
        }
        if (!this.isValidRawTransaction(rawTransaction)) {
            throw new sdk_core_1.ParseTransactionError('Invalid raw transaction');
        }
    }
    /** @inheritdoc */
    validateValue(value) {
        if (value.isNaN()) {
            throw new sdk_core_1.BuildTransactionError('Invalid amount format');
        }
        else if (value.isLessThan(0)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be less than zero');
        }
    }
    validateGasData(gasData) {
        this.validateValue(new bignumber_js_1.default(gasData.maxGasAmount));
        this.validateValue(new bignumber_js_1.default(gasData.gasUnitPrice));
    }
    addFeePayerAddress(address) {
        this.transaction.addFeePayerAddress(address);
    }
    getFeePayerPubKey() {
        return this.transaction.getFeePayerPubKey();
    }
}
exports.TransactionBuilder = TransactionBuilder;
