// @bitgo/public-types/dist/src/schema/mpcv2/smc/messages/p2p.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthEncP2PMessage = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const common_1 = require("../../common");
const AuthEncMessage = t.type({
    encryptedMessage: io_ts_types_1.NonEmptyString,
    signature: io_ts_types_1.NonEmptyString,
});
exports.AuthEncP2PMessage = t.intersection([
    t.type({
        payload: AuthEncMessage,
        from: common_1.MPCv2PartyFromStringOrNumber,
        to: common_1.MPCv2PartyFromStringOrNumber,
    }),
    t.partial({
        commitment: io_ts_types_1.NonEmptyString,
    }),
]);
