// @taquito/rpc/dist/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartRollupRefuteGameEndedReason = exports.SmartRollupRefuteGameEndedPlayerOutcomes = exports.SmartRollupRefuteGameStatusOptions = exports.SmartRollupRefuteRevealProofKind = exports.SmartRollupInputProofKind = exports.SmartRollupRefutationOptions = exports.PvmKind = exports.METADATA_BALANCE_UPDATES_CATEGORY = exports.OPERATION_METADATA = void 0;
var OPERATION_METADATA;
(function (OPERATION_METADATA) {
    OPERATION_METADATA["TOO_LARGE"] = "too large";
})(OPERATION_METADATA || (exports.OPERATION_METADATA = OPERATION_METADATA = {}));
var METADATA_BALANCE_UPDATES_CATEGORY;
(function (METADATA_BALANCE_UPDATES_CATEGORY) {
    METADATA_BALANCE_UPDATES_CATEGORY["ACTIVATION"] = "activation";
    METADATA_BALANCE_UPDATES_CATEGORY["ATTESTATION"] = "attestation";
    METADATA_BALANCE_UPDATES_CATEGORY["ATTESTATIONS_AGGREGATE"] = "attestations_aggregate";
    METADATA_BALANCE_UPDATES_CATEGORY["ATTESTATION_WITH_DAL"] = "attestation_with_dal";
    METADATA_BALANCE_UPDATES_CATEGORY["ATTESTING_REWARDS"] = "attesting rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["BAKING_BONUSES"] = "baking bonuses";
    METADATA_BALANCE_UPDATES_CATEGORY["BAKING_REWARDS"] = "baking rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["BALLOT"] = "ballot";
    METADATA_BALANCE_UPDATES_CATEGORY["BLOCK_FEES"] = "block fees";
    METADATA_BALANCE_UPDATES_CATEGORY["BONDS"] = "bonds";
    METADATA_BALANCE_UPDATES_CATEGORY["BOOTSTRAP"] = "bootstrap";
    METADATA_BALANCE_UPDATES_CATEGORY["BURNED"] = "burned";
    METADATA_BALANCE_UPDATES_CATEGORY["COMMITMENT"] = "commitment";
    METADATA_BALANCE_UPDATES_CATEGORY["DAL_ATTESTING_REWARDS"] = "dal attesting rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["DAL_ENTRAPMENT_EVIDENCE"] = "dal_entrapment_evidence";
    METADATA_BALANCE_UPDATES_CATEGORY["DAL_PUBLISH_COMMITMENT"] = "dal_publish_commitment";
    METADATA_BALANCE_UPDATES_CATEGORY["DELEGATE_DENOMINATOR"] = "delegate_denominator";
    METADATA_BALANCE_UPDATES_CATEGORY["DELEGATION"] = "delegation";
    METADATA_BALANCE_UPDATES_CATEGORY["DELEGATOR_NUMERATOR"] = "delegator_numerator";
    METADATA_BALANCE_UPDATES_CATEGORY["DEPOSITS"] = "deposits";
    METADATA_BALANCE_UPDATES_CATEGORY["DOUBLE_ATTESTATION_EVIDENCE"] = "double_attestation_evidence";
    METADATA_BALANCE_UPDATES_CATEGORY["DOUBLE_BAKING_EVIDENCE"] = "double_baking_evidence";
    METADATA_BALANCE_UPDATES_CATEGORY["DOUBLE_ENDORSEMENT_EVIDENCE"] = "double_endorsement_evidence";
    METADATA_BALANCE_UPDATES_CATEGORY["DOUBLE_PREATTESTATION_EVIDENCE"] = "double_preattestation_evidence";
    METADATA_BALANCE_UPDATES_CATEGORY["DOUBLE_PREENDORSEMENT_EVIDENCE"] = "double_preendorsement_evidence";
    METADATA_BALANCE_UPDATES_CATEGORY["DRAIN_DELEGATE"] = "drain_delegate";
    METADATA_BALANCE_UPDATES_CATEGORY["ENDORSEMENT"] = "endorsement";
    METADATA_BALANCE_UPDATES_CATEGORY["ENDORSEMENT_WITH_DAL"] = "endorsement_with_dal";
    METADATA_BALANCE_UPDATES_CATEGORY["EVENT"] = "event";
    METADATA_BALANCE_UPDATES_CATEGORY["FAILING_NOOP"] = "failing_noop";
    METADATA_BALANCE_UPDATES_CATEGORY["INCREASE_PAID_STORAGE"] = "increase_paid_storage";
    METADATA_BALANCE_UPDATES_CATEGORY["INVOICE"] = "invoice";
    METADATA_BALANCE_UPDATES_CATEGORY["LOST_ATTESTING_REWARDS"] = "lost attesting rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["LOST_DAL_ATTESTING_REWARDS"] = "lost dal attesting rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["MINTED"] = "minted";
    METADATA_BALANCE_UPDATES_CATEGORY["NONCE_REVELATION_REWARDS"] = "nonce revelation rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["ORIGINATION"] = "origination";
    METADATA_BALANCE_UPDATES_CATEGORY["PREATTESTATION"] = "preattestation";
    METADATA_BALANCE_UPDATES_CATEGORY["PREATTESTATIONS_AGGREGATE"] = "preattestations_aggregate";
    METADATA_BALANCE_UPDATES_CATEGORY["PREENDORSEMENT"] = "preendorsement";
    METADATA_BALANCE_UPDATES_CATEGORY["PROPOSALS"] = "proposals";
    METADATA_BALANCE_UPDATES_CATEGORY["PUNISHMENTS"] = "punishments";
    METADATA_BALANCE_UPDATES_CATEGORY["REGISTER_GLOBAL_CONSTANT"] = "register_global_constant";
    METADATA_BALANCE_UPDATES_CATEGORY["REVEAL"] = "reveal";
    METADATA_BALANCE_UPDATES_CATEGORY["SEED_NONCE_REVELATION"] = "seed_nonce_revelation";
    METADATA_BALANCE_UPDATES_CATEGORY["SET_DEPOSITS_LIMIT"] = "set_deposits_limit";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_ADD_MESSAGES"] = "smart_rollup_add_messages";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_CEMENT"] = "smart_rollup_cement";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_EXECUTE_OUTBOX_MESSAGE"] = "smart_rollup_execute_outbox_message";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_ORIGINATE"] = "smart_rollup_originate";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_PUBLISH"] = "smart_rollup_publish";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_RECOVER_BOND"] = "smart_rollup_recover_bond";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_REFUTATION_PUNISHMENTS"] = "smart_rollup_refutation_punishments";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_REFUTATION_REWARDS"] = "smart_rollup_refutation_rewards";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_REFUTE"] = "smart_rollup_refute";
    METADATA_BALANCE_UPDATES_CATEGORY["SMART_ROLLUP_TIMEOUT"] = "smart_rollup_timeout";
    METADATA_BALANCE_UPDATES_CATEGORY["STORAGE_FEES"] = "storage fees";
    METADATA_BALANCE_UPDATES_CATEGORY["SUBSIDY"] = "subsidy";
    METADATA_BALANCE_UPDATES_CATEGORY["TICKET_UPDATES"] = "ticket_updates";
    METADATA_BALANCE_UPDATES_CATEGORY["TRANSACTION"] = "transaction";
    METADATA_BALANCE_UPDATES_CATEGORY["TRANSFER_TICKET"] = "transfer_ticket";
    METADATA_BALANCE_UPDATES_CATEGORY["UNSTAKED_DEPOSITS"] = "unstaked_deposits";
    METADATA_BALANCE_UPDATES_CATEGORY["UPDATE_COMPANION_KEY"] = "update_companion_key";
    METADATA_BALANCE_UPDATES_CATEGORY["UPDATE_CONSENSUS_KEY"] = "update_consensus_key";
    METADATA_BALANCE_UPDATES_CATEGORY["VDF_REVELATION"] = "vdf_revelation";
})(METADATA_BALANCE_UPDATES_CATEGORY || (exports.METADATA_BALANCE_UPDATES_CATEGORY = METADATA_BALANCE_UPDATES_CATEGORY = {}));
var PvmKind;
(function (PvmKind) {
    PvmKind["WASM2"] = "wasm_2_0_0";
    PvmKind["ARITH"] = "arith";
    PvmKind["RISCV"] = "riscv";
})(PvmKind || (exports.PvmKind = PvmKind = {}));
var SmartRollupRefutationOptions;
(function (SmartRollupRefutationOptions) {
    SmartRollupRefutationOptions["START"] = "start";
    SmartRollupRefutationOptions["MOVE"] = "move";
})(SmartRollupRefutationOptions || (exports.SmartRollupRefutationOptions = SmartRollupRefutationOptions = {}));
var SmartRollupInputProofKind;
(function (SmartRollupInputProofKind) {
    SmartRollupInputProofKind["INBOX_PROOF"] = "inbox_proof";
    SmartRollupInputProofKind["REVEAL_PROOF"] = "reveal_proof";
    SmartRollupInputProofKind["FIRST_INPUT"] = "first_input";
})(SmartRollupInputProofKind || (exports.SmartRollupInputProofKind = SmartRollupInputProofKind = {}));
var SmartRollupRefuteRevealProofKind;
(function (SmartRollupRefuteRevealProofKind) {
    SmartRollupRefuteRevealProofKind["RAW_DATA_PROOF"] = "raw_data_proof";
    SmartRollupRefuteRevealProofKind["METADATA_PROOF"] = "metadata_proof";
    SmartRollupRefuteRevealProofKind["DAL_PAGE_PROOF"] = "dal_page_proof";
    SmartRollupRefuteRevealProofKind["DAL_PARAMETERS_PROOF"] = "dal_parameters_proof";
})(SmartRollupRefuteRevealProofKind || (exports.SmartRollupRefuteRevealProofKind = SmartRollupRefuteRevealProofKind = {}));
var SmartRollupRefuteGameStatusOptions;
(function (SmartRollupRefuteGameStatusOptions) {
    SmartRollupRefuteGameStatusOptions["ONGOING"] = "ongoing";
    SmartRollupRefuteGameStatusOptions["ENDED"] = "ended";
})(SmartRollupRefuteGameStatusOptions || (exports.SmartRollupRefuteGameStatusOptions = SmartRollupRefuteGameStatusOptions = {}));
var SmartRollupRefuteGameEndedPlayerOutcomes;
(function (SmartRollupRefuteGameEndedPlayerOutcomes) {
    SmartRollupRefuteGameEndedPlayerOutcomes["LOSER"] = "loser";
    SmartRollupRefuteGameEndedPlayerOutcomes["DRAW"] = "draw";
})(SmartRollupRefuteGameEndedPlayerOutcomes || (exports.SmartRollupRefuteGameEndedPlayerOutcomes = SmartRollupRefuteGameEndedPlayerOutcomes = {}));
var SmartRollupRefuteGameEndedReason;
(function (SmartRollupRefuteGameEndedReason) {
    SmartRollupRefuteGameEndedReason["CONFLICT_RESOLVED"] = "conflict_resolved";
    SmartRollupRefuteGameEndedReason["TIMEOUT"] = "timeout";
})(SmartRollupRefuteGameEndedReason || (exports.SmartRollupRefuteGameEndedReason = SmartRollupRefuteGameEndedReason = {}));
